// picomatch/lib/utils.js
'use strict';

const path = require('path');
const win32 = process.platform === 'win32';
const {
  REGEX_BACKSLASH,
  REGEX_REMOVE_BACKSLASH,
  REGEX_SPECIAL_CHARS,
  REGEX_SPECIAL_CHARS_GLOBAL
***REMOVED*** = require('./constants');

exports.isObject = val => val !== null && typeof val === 'object' && !Array.isArray(val);
exports.hasRegexChars = str => REGEX_SPECIAL_CHARS.test(str);
exports.isRegexChar = str => str.length === 1 && exports.hasRegexChars(str);
exports.escapeRegex = str => str.replace(REGEX_SPECIAL_CHARS_GLOBAL, '\\$1');
exports.toPosixSlashes = str => str.replace(REGEX_BACKSLASH, '/');

exports.removeBackslashes = str => {
  return str.replace(REGEX_REMOVE_BACKSLASH, match => {
    return match === '\\' ? '' : match;
  ***REMOVED***);
***REMOVED***;

exports.supportsLookbehinds = () => {
  const segs = process.version.slice(1).split('.').map(Number);
  if (segs.length === 3 && segs[0] >= 9 || (segs[0] === 8 && segs[1] >= 10)) {
    return true;
  ***REMOVED***
  return false;
***REMOVED***;

exports.isWindows = options => {
  if (options && typeof options.windows === 'boolean') {
    return options.windows;
  ***REMOVED***
  return win32 === true || path.sep === '\\';
***REMOVED***;

exports.escapeLast = (input, char, lastIdx) => {
  const idx = input.lastIndexOf(char, lastIdx);
  if (idx === -1) return input;
  if (input[idx - 1] === '\\') return exports.escapeLast(input, char, idx - 1);
  return `${input.slice(0, idx)***REMOVED***\\${input.slice(idx)***REMOVED***`;
***REMOVED***;

exports.removePrefix = (input, state = {***REMOVED***) => {
  let output = input;
  if (output.startsWith('./')) {
    output = output.slice(2);
    state.prefix = './';
  ***REMOVED***
  return output;
***REMOVED***;

exports.wrapOutput = (input, state = {***REMOVED***, options = {***REMOVED***) => {
  const prepend = options.contains ? '' : '^';
  const append = options.contains ? '' : '$';

  let output = `${prepend***REMOVED***(?:${input***REMOVED***)${append***REMOVED***`;
  if (state.negated === true) {
    output = `(?:^(?!${output***REMOVED***).*$)`;
  ***REMOVED***
  return output;
***REMOVED***;
