// @jest/core/build/plugins/FailedTestsInteractive.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
***REMOVED***);
exports.default = void 0;

function _jestWatcher() {
  const data = require('jest-watcher');

  _jestWatcher = function () {
    return data;
  ***REMOVED***;

  return data;
***REMOVED***

var _FailedTestsInteractiveMode = _interopRequireDefault(
  require('../FailedTestsInteractiveMode')
);

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {default: obj***REMOVED***;
***REMOVED***

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    ***REMOVED***);
  ***REMOVED*** else {
    obj[key] = value;
  ***REMOVED***
  return obj;
***REMOVED***

class FailedTestsInteractivePlugin extends _jestWatcher().BaseWatchPlugin {
  constructor(...args) {
    super(...args);

    _defineProperty(this, '_failedTestAssertions', void 0);

    _defineProperty(
      this,
      '_manager',
      new _FailedTestsInteractiveMode.default(this._stdout)
    );
  ***REMOVED***

  apply(hooks) {
    hooks.onTestRunComplete(results => {
      this._failedTestAssertions = this.getFailedTestAssertions(results);
      if (this._manager.isActive()) this._manager.updateWithResults(results);
    ***REMOVED***);
  ***REMOVED***

  getUsageInfo() {
    var _this$_failedTestAsse;

    if (
      (_this$_failedTestAsse = this._failedTestAssertions) !== null &&
      _this$_failedTestAsse !== void 0 &&
      _this$_failedTestAsse.length
    ) {
      return {
        key: 'i',
        prompt: 'run failing tests interactively'
      ***REMOVED***;
    ***REMOVED***

    return null;
  ***REMOVED***

  onKey(key) {
    if (this._manager.isActive()) {
      this._manager.put(key);
    ***REMOVED***
  ***REMOVED***

  run(_, updateConfigAndRun) {
    return new Promise(resolve => {
      if (
        !this._failedTestAssertions ||
        this._failedTestAssertions.length === 0
      ) {
        resolve();
        return;
      ***REMOVED***

      this._manager.run(this._failedTestAssertions, failure => {
        updateConfigAndRun({
          mode: 'watch',
          testNamePattern: failure ? `^${failure.fullName***REMOVED***$` : '',
          testPathPattern:
            (failure === null || failure === void 0 ? void 0 : failure.path) ||
            ''
        ***REMOVED***);

        if (!this._manager.isActive()) {
          resolve();
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***

  getFailedTestAssertions(results) {
    const failedTestPaths = [];

    if (
      // skip if no failed tests
      results.numFailedTests === 0 || // skip if missing test results
      !results.testResults || // skip if unmatched snapshots are present
      results.snapshot.unmatched
    ) {
      return failedTestPaths;
    ***REMOVED***

    results.testResults.forEach(testResult => {
      testResult.testResults.forEach(result => {
        if (result.status === 'failed') {
          failedTestPaths.push({
            fullName: result.fullName,
            path: testResult.testFilePath
          ***REMOVED***);
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***);
    return failedTestPaths;
  ***REMOVED***
***REMOVED***

exports.default = FailedTestsInteractivePlugin;
