// @jest/core/build/plugins/TestPathPattern.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
***REMOVED***);
exports.default = void 0;

function _jestWatcher() {
  const data = require('jest-watcher');

  _jestWatcher = function () {
    return data;
  ***REMOVED***;

  return data;
***REMOVED***

var _TestPathPatternPrompt = _interopRequireDefault(
  require('../TestPathPatternPrompt')
);

var _activeFiltersMessage = _interopRequireDefault(
  require('../lib/activeFiltersMessage')
);

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {default: obj***REMOVED***;
***REMOVED***

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    ***REMOVED***);
  ***REMOVED*** else {
    obj[key] = value;
  ***REMOVED***
  return obj;
***REMOVED***

class TestPathPatternPlugin extends _jestWatcher().BaseWatchPlugin {
  constructor(options) {
    super(options);

    _defineProperty(this, '_prompt', void 0);

    _defineProperty(this, 'isInternal', void 0);

    this._prompt = new (_jestWatcher().Prompt)();
    this.isInternal = true;
  ***REMOVED***

  getUsageInfo() {
    return {
      key: 'p',
      prompt: 'filter by a filename regex pattern'
    ***REMOVED***;
  ***REMOVED***

  onKey(key) {
    this._prompt.put(key);
  ***REMOVED***

  run(globalConfig, updateConfigAndRun) {
    return new Promise((res, rej) => {
      const testPathPatternPrompt = new _TestPathPatternPrompt.default(
        this._stdout,
        this._prompt
      );
      testPathPatternPrompt.run(
        value => {
          updateConfigAndRun({
            mode: 'watch',
            testPathPattern: value
          ***REMOVED***);
          res();
        ***REMOVED***,
        rej,
        {
          header: (0, _activeFiltersMessage.default)(globalConfig)
        ***REMOVED***
      );
    ***REMOVED***);
  ***REMOVED***
***REMOVED***

var _default = TestPathPatternPlugin;
exports.default = _default;
