// @sinclair/typebox/typebox.js
"use strict";
/*--------------------------------------------------------------------------

@sinclair/typebox

The MIT License (MIT)

Copyright (c) 2022 Haydn Paterson (sinclair) <haydn.developer@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

---------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.Type = exports.TypeBuilder = exports.Modifier = exports.Hint = exports.Kind = void 0;
// --------------------------------------------------------------------------
// Symbols
// --------------------------------------------------------------------------
exports.Kind = Symbol.for('TypeBox.Kind');
exports.Hint = Symbol.for('TypeBox.Hint');
exports.Modifier = Symbol.for('TypeBox.Modifier');
// --------------------------------------------------------------------------
// TypeBuilder
// --------------------------------------------------------------------------
let TypeOrdinal = 0;
class TypeBuilder {
    // ----------------------------------------------------------------------
    // Modifiers
    // ----------------------------------------------------------------------
    /** Creates a readonly optional property */
    ReadonlyOptional(item) {
        return { [exports.Modifier]: 'ReadonlyOptional', ...item ***REMOVED***;
    ***REMOVED***
    /** Creates a readonly property */
    Readonly(item) {
        return { [exports.Modifier]: 'Readonly', ...item ***REMOVED***;
    ***REMOVED***
    /** Creates a optional property */
    Optional(item) {
        return { [exports.Modifier]: 'Optional', ...item ***REMOVED***;
    ***REMOVED***
    // ----------------------------------------------------------------------
    // Types
    // ----------------------------------------------------------------------
    /** Creates a any type */
    Any(options = {***REMOVED***) {
        return this.Create({ ...options, [exports.Kind]: 'Any' ***REMOVED***);
    ***REMOVED***
    /** Creates a array type */
    Array(items, options = {***REMOVED***) {
        return this.Create({ ...options, [exports.Kind]: 'Array', type: 'array', items ***REMOVED***);
    ***REMOVED***
    /** Creates a boolean type */
    Boolean(options = {***REMOVED***) {
        return this.Create({ ...options, [exports.Kind]: 'Boolean', type: 'boolean' ***REMOVED***);
    ***REMOVED***
    /** Creates a tuple type from this constructors parameters */
    ConstructorParameters(schema, options = {***REMOVED***) {
        return this.Tuple([...schema.parameters], { ...options ***REMOVED***);
    ***REMOVED***
    /** Creates a constructor type */
    Constructor(parameters, returns, options = {***REMOVED***) {
        if (parameters[exports.Kind] === 'Tuple') {
            const inner = parameters.items === undefined ? [] : parameters.items;
            return this.Create({ ...options, [exports.Kind]: 'Constructor', type: 'constructor', parameters: inner, returns ***REMOVED***);
        ***REMOVED***
        else if (globalThis.Array.isArray(parameters)) {
            return this.Create({ ...options, [exports.Kind]: 'Constructor', type: 'constructor', parameters, returns ***REMOVED***);
        ***REMOVED***
        else {
            throw new Error('TypeBuilder.Constructor: Invalid parameters');
        ***REMOVED***
    ***REMOVED***
    /** Creates a enum type */
    Enum(item, options = {***REMOVED***) {
        const values = Object.keys(item)
            .filter((key) => isNaN(key))
            .map((key) => item[key]);
        const anyOf = values.map((value) => (typeof value === 'string' ? { [exports.Kind]: 'Literal', type: 'string', const: value ***REMOVED*** : { [exports.Kind]: 'Literal', type: 'number', const: value ***REMOVED***));
        return this.Create({ ...options, [exports.Kind]: 'Union', [exports.Hint]: 'Enum', anyOf ***REMOVED***);
    ***REMOVED***
    /** Creates a function type */
    Function(parameters, returns, options = {***REMOVED***) {
        if (parameters[exports.Kind] === 'Tuple') {
            const inner = parameters.items === undefined ? [] : parameters.items;
            return this.Create({ ...options, [exports.Kind]: 'Function', type: 'function', parameters: inner, returns ***REMOVED***);
        ***REMOVED***
        else if (globalThis.Array.isArray(parameters)) {
            return this.Create({ ...options, [exports.Kind]: 'Function', type: 'function', parameters, returns ***REMOVED***);
        ***REMOVED***
        else {
            throw new Error('TypeBuilder.Function: Invalid parameters');
        ***REMOVED***
    ***REMOVED***
    /** Creates a type from this constructors instance type */
    InstanceType(schema, options = {***REMOVED***) {
        return { ...options, ...this.Clone(schema.returns) ***REMOVED***;
    ***REMOVED***
    /** Creates a integer type */
    Integer(options = {***REMOVED***) {
        return this.Create({ ...options, [exports.Kind]: 'Integer', type: 'integer' ***REMOVED***);
    ***REMOVED***
    /** Creates a intersect type. */
    Intersect(objects, options = {***REMOVED***) {
        const isOptional = (schema) => (schema[exports.Modifier] && schema[exports.Modifier] === 'Optional') || schema[exports.Modifier] === 'ReadonlyOptional';
        const [required, optional] = [new Set(), new Set()];
        for (const object of objects) {
            for (const [key, schema] of Object.entries(object.properties)) {
                if (isOptional(schema))
                    optional.add(key);
            ***REMOVED***
        ***REMOVED***
        for (const object of objects) {
            for (const key of Object.keys(object.properties)) {
                if (!optional.has(key))
                    required.add(key);
            ***REMOVED***
        ***REMOVED***
        const properties = {***REMOVED***;
        for (const object of objects) {
            for (const [key, schema] of Object.entries(object.properties)) {
                properties[key] = properties[key] === undefined ? schema : { [exports.Kind]: 'Union', anyOf: [properties[key], { ...schema ***REMOVED***] ***REMOVED***;
            ***REMOVED***
        ***REMOVED***
        if (required.size > 0) {
            return this.Create({ ...options, [exports.Kind]: 'Object', type: 'object', properties, required: [...required] ***REMOVED***);
        ***REMOVED***
        else {
            return this.Create({ ...options, [exports.Kind]: 'Object', type: 'object', properties ***REMOVED***);
        ***REMOVED***
    ***REMOVED***
    /** Creates a keyof type */
    KeyOf(object, options = {***REMOVED***) {
        const items = Object.keys(object.properties).map((key) => this.Create({ ...options, [exports.Kind]: 'Literal', type: 'string', const: key ***REMOVED***));
        return this.Create({ ...options, [exports.Kind]: 'Union', [exports.Hint]: 'KeyOf', anyOf: items ***REMOVED***);
    ***REMOVED***
    /** Creates a literal type. */
    Literal(value, options = {***REMOVED***) {
        return this.Create({ ...options, [exports.Kind]: 'Literal', const: value, type: typeof value ***REMOVED***);
    ***REMOVED***
    /** Creates a never type */
    Never(options = {***REMOVED***) {
        return this.Create({
            ...options,
            [exports.Kind]: 'Never',
            allOf: [
                { type: 'boolean', const: false ***REMOVED***,
                { type: 'boolean', const: true ***REMOVED***,
            ],
        ***REMOVED***);
    ***REMOVED***
    /** Creates a null type */
    Null(options = {***REMOVED***) {
        return this.Create({ ...options, [exports.Kind]: 'Null', type: 'null' ***REMOVED***);
    ***REMOVED***
    /** Creates a number type */
    Number(options = {***REMOVED***) {
        return this.Create({ ...options, [exports.Kind]: 'Number', type: 'number' ***REMOVED***);
    ***REMOVED***
    /** Creates an object type with the given properties */
    Object(properties, options = {***REMOVED***) {
        const property_names = Object.keys(properties);
        const optional = property_names.filter((name) => {
            const property = properties[name];
            const modifier = property[exports.Modifier];
            return modifier && (modifier === 'Optional' || modifier === 'ReadonlyOptional');
        ***REMOVED***);
        const required = property_names.filter((name) => !optional.includes(name));
        if (required.length > 0) {
            return this.Create({ ...options, [exports.Kind]: 'Object', type: 'object', properties, required ***REMOVED***);
        ***REMOVED***
        else {
            return this.Create({ ...options, [exports.Kind]: 'Object', type: 'object', properties ***REMOVED***);
        ***REMOVED***
    ***REMOVED***
    /** Creates a new object whose properties are omitted from the given object */
    Omit(schema, keys, options = {***REMOVED***) {
        const select = keys[exports.Kind] === 'Union' ? keys.anyOf.map((schema) => schema.const) : keys;
        const next = { ...this.Clone(schema), ...options, [exports.Hint]: 'Omit' ***REMOVED***;
        if (next.required) {
            next.required = next.required.filter((key) => !select.includes(key));
            if (next.required.length === 0)
                delete next.required;
        ***REMOVED***
        for (const key of Object.keys(next.properties)) {
            if (select.includes(key))
                delete next.properties[key];
        ***REMOVED***
        return this.Create(next);
    ***REMOVED***
    /** Creates a tuple type from this functions parameters */
    Parameters(schema, options = {***REMOVED***) {
        return exports.Type.Tuple(schema.parameters, { ...options ***REMOVED***);
    ***REMOVED***
    /** Creates an object type whose properties are all optional */
    Partial(schema, options = {***REMOVED***) {
        const next = { ...this.Clone(schema), ...options, [exports.Hint]: 'Partial' ***REMOVED***;
        delete next.required;
        for (const key of Object.keys(next.properties)) {
            const property = next.properties[key];
            const modifer = property[exports.Modifier];
            switch (modifer) {
                case 'ReadonlyOptional':
                    property[exports.Modifier] = 'ReadonlyOptional';
                    break;
                case 'Readonly':
                    property[exports.Modifier] = 'ReadonlyOptional';
                    break;
                case 'Optional':
                    property[exports.Modifier] = 'Optional';
                    break;
                default:
                    property[exports.Modifier] = 'Optional';
                    break;
            ***REMOVED***
        ***REMOVED***
        return this.Create(next);
    ***REMOVED***
    /** Creates a object whose properties are picked from the given object */
    Pick(schema, keys, options = {***REMOVED***) {
        const select = keys[exports.Kind] === 'Union' ? keys.anyOf.map((schema) => schema.const) : keys;
        const next = { ...this.Clone(schema), ...options, [exports.Hint]: 'Pick' ***REMOVED***;
        if (next.required) {
            next.required = next.required.filter((key) => select.includes(key));
            if (next.required.length === 0)
                delete next.required;
        ***REMOVED***
        for (const key of Object.keys(next.properties)) {
            if (!select.includes(key))
                delete next.properties[key];
        ***REMOVED***
        return this.Create(next);
    ***REMOVED***
    /** Creates a promise type. This type cannot be represented in schema. */
    Promise(item, options = {***REMOVED***) {
        return this.Create({ ...options, [exports.Kind]: 'Promise', type: 'promise', item ***REMOVED***);
    ***REMOVED***
    /** Creates a record type */
    Record(key, value, options = {***REMOVED***) {
        // If string literal union return TObject with properties extracted from union.
        if (key[exports.Kind] === 'Union') {
            return this.Object(key.anyOf.reduce((acc, literal) => {
                return { ...acc, [literal.const]: value ***REMOVED***;
            ***REMOVED***, {***REMOVED***), { ...options, [exports.Hint]: 'Record' ***REMOVED***);
        ***REMOVED***
        // otherwise return TRecord with patternProperties
        const pattern = ['Integer', 'Number'].includes(key[exports.Kind]) ? '^(0|[1-9][0-9]*)$' : key[exports.Kind] === 'String' && key.pattern ? key.pattern : '^.*$';
        return this.Create({
            ...options,
            [exports.Kind]: 'Record',
            type: 'object',
            patternProperties: { [pattern]: value ***REMOVED***,
            additionalProperties: false,
        ***REMOVED***);
    ***REMOVED***
    /** Creates a recursive object type */
    Recursive(callback, options = {***REMOVED***) {
        if (options.$id === undefined)
            options.$id = `T${TypeOrdinal++***REMOVED***`;
        const self = callback({ [exports.Kind]: 'Self', $ref: `${options.$id***REMOVED***` ***REMOVED***);
        self.$id = options.$id;
        return this.Create({ ...options, ...self ***REMOVED***);
    ***REMOVED***
    /** Creates a reference schema */
    Ref(schema, options = {***REMOVED***) {
        if (schema.$id === undefined)
            throw Error('TypeBuilder.Ref: Referenced schema must specify an $id');
        return this.Create({ ...options, [exports.Kind]: 'Ref', $ref: schema.$id ***REMOVED***);
    ***REMOVED***
    /** Creates a string type from a regular expression */
    RegEx(regex, options = {***REMOVED***) {
        return this.Create({ ...options, [exports.Kind]: 'String', type: 'string', pattern: regex.source ***REMOVED***);
    ***REMOVED***
    /** Creates an object type whose properties are all required */
    Required(schema, options = {***REMOVED***) {
        const next = { ...this.Clone(schema), ...options, [exports.Hint]: 'Required' ***REMOVED***;
        next.required = Object.keys(next.properties);
        for (const key of Object.keys(next.properties)) {
            const property = next.properties[key];
            const modifier = property[exports.Modifier];
            switch (modifier) {
                case 'ReadonlyOptional':
                    property[exports.Modifier] = 'Readonly';
                    break;
                case 'Readonly':
                    property[exports.Modifier] = 'Readonly';
                    break;
                case 'Optional':
                    delete property[exports.Modifier];
                    break;
                default:
                    delete property[exports.Modifier];
                    break;
            ***REMOVED***
        ***REMOVED***
        return this.Create(next);
    ***REMOVED***
    /** Creates a type from this functions return type */
    ReturnType(schema, options = {***REMOVED***) {
        return { ...options, ...this.Clone(schema.returns) ***REMOVED***;
    ***REMOVED***
    /** Removes Kind and Modifier symbol property keys from this schema */
    Strict(schema) {
        return JSON.parse(JSON.stringify(schema));
    ***REMOVED***
    /** Creates a string type */
    String(options = {***REMOVED***) {
        return this.Create({ ...options, [exports.Kind]: 'String', type: 'string' ***REMOVED***);
    ***REMOVED***
    /** Creates a tuple type */
    Tuple(items, options = {***REMOVED***) {
        const additionalItems = false;
        const minItems = items.length;
        const maxItems = items.length;
        const schema = (items.length > 0 ? { ...options, [exports.Kind]: 'Tuple', type: 'array', items, additionalItems, minItems, maxItems ***REMOVED*** : { ...options, [exports.Kind]: 'Tuple', type: 'array', minItems, maxItems ***REMOVED***);
        return this.Create(schema);
    ***REMOVED***
    /** Creates a undefined type */
    Undefined(options = {***REMOVED***) {
        return this.Create({ ...options, [exports.Kind]: 'Undefined', type: 'object', specialized: 'Undefined' ***REMOVED***);
    ***REMOVED***
    Union(items, options = {***REMOVED***) {
        return items.length === 0 ? exports.Type.Never({ ...options ***REMOVED***) : this.Create({ ...options, [exports.Kind]: 'Union', anyOf: items ***REMOVED***);
    ***REMOVED***
    /** Creates a Uint8Array type */
    Uint8Array(options = {***REMOVED***) {
        return this.Create({ ...options, [exports.Kind]: 'Uint8Array', type: 'object', specialized: 'Uint8Array' ***REMOVED***);
    ***REMOVED***
    /** Creates an unknown type */
    Unknown(options = {***REMOVED***) {
        return this.Create({ ...options, [exports.Kind]: 'Unknown' ***REMOVED***);
    ***REMOVED***
    /** Creates a user defined schema that infers as type T  */
    Unsafe(options = {***REMOVED***) {
        return this.Create({ ...options, [exports.Kind]: options[exports.Kind] || 'Unsafe' ***REMOVED***);
    ***REMOVED***
    /** Creates a void type */
    Void(options = {***REMOVED***) {
        return this.Create({ ...options, [exports.Kind]: 'Void', type: 'null' ***REMOVED***);
    ***REMOVED***
    /** Use this function to return TSchema with static and params omitted */
    Create(schema) {
        return schema;
    ***REMOVED***
    /** Clones the given value */
    Clone(value) {
        const isObject = (object) => typeof object === 'object' && object !== null && !Array.isArray(object);
        const isArray = (object) => typeof object === 'object' && object !== null && Array.isArray(object);
        if (isObject(value)) {
            return Object.keys(value).reduce((acc, key) => ({
                ...acc,
                [key]: this.Clone(value[key]),
            ***REMOVED***), Object.getOwnPropertySymbols(value).reduce((acc, key) => ({
                ...acc,
                [key]: this.Clone(value[key]),
            ***REMOVED***), {***REMOVED***));
        ***REMOVED***
        else if (isArray(value)) {
            return value.map((item) => this.Clone(item));
        ***REMOVED***
        else {
            return value;
        ***REMOVED***
    ***REMOVED***
***REMOVED***
exports.TypeBuilder = TypeBuilder;
/** JSON Schema Type Builder with Static Type Resolution for TypeScript */
exports.Type = new TypeBuilder();
