// webpack/lib/node/NodeWatchFileSystem.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const util = require("util");
const Watchpack = require("watchpack");

/** @typedef {import("../../declarations/WebpackOptions").WatchOptions***REMOVED*** WatchOptions */
/** @typedef {import("../FileSystemInfo").FileSystemInfoEntry***REMOVED*** FileSystemInfoEntry */
/** @typedef {import("../util/fs").InputFileSystem***REMOVED*** InputFileSystem */
/** @typedef {import("../util/fs").WatchMethod***REMOVED*** WatchMethod */

class NodeWatchFileSystem {
	/**
	 * @param {InputFileSystem***REMOVED*** inputFileSystem input filesystem
	 */
	constructor(inputFileSystem) {
		this.inputFileSystem = inputFileSystem;
		this.watcherOptions = {
			aggregateTimeout: 0
		***REMOVED***;
		/** @type {Watchpack | null***REMOVED*** */
		this.watcher = new Watchpack(this.watcherOptions);
	***REMOVED***

	/** @type {WatchMethod***REMOVED*** */
	watch(
		files,
		directories,
		missing,
		startTime,
		options,
		callback,
		callbackUndelayed
	) {
		if (!files || typeof files[Symbol.iterator] !== "function") {
			throw new Error("Invalid arguments: 'files'");
		***REMOVED***
		if (!directories || typeof directories[Symbol.iterator] !== "function") {
			throw new Error("Invalid arguments: 'directories'");
		***REMOVED***
		if (!missing || typeof missing[Symbol.iterator] !== "function") {
			throw new Error("Invalid arguments: 'missing'");
		***REMOVED***
		if (typeof callback !== "function") {
			throw new Error("Invalid arguments: 'callback'");
		***REMOVED***
		if (typeof startTime !== "number" && startTime) {
			throw new Error("Invalid arguments: 'startTime'");
		***REMOVED***
		if (typeof options !== "object") {
			throw new Error("Invalid arguments: 'options'");
		***REMOVED***
		if (typeof callbackUndelayed !== "function" && callbackUndelayed) {
			throw new Error("Invalid arguments: 'callbackUndelayed'");
		***REMOVED***
		const oldWatcher = this.watcher;
		this.watcher = new Watchpack(options);

		if (callbackUndelayed) {
			this.watcher.once("change", callbackUndelayed);
		***REMOVED***

		const fetchTimeInfo = () => {
			const fileTimeInfoEntries = new Map();
			const contextTimeInfoEntries = new Map();
			if (this.watcher) {
				this.watcher.collectTimeInfoEntries(
					fileTimeInfoEntries,
					contextTimeInfoEntries
				);
			***REMOVED***
			return { fileTimeInfoEntries, contextTimeInfoEntries ***REMOVED***;
		***REMOVED***;
		this.watcher.once(
			"aggregated",
			/**
			 * @param {Set<string>***REMOVED*** changes changes
			 * @param {Set<string>***REMOVED*** removals removals
			 */
			(changes, removals) => {
				// pause emitting events (avoids clearing aggregated changes and removals on timeout)
				/** @type {Watchpack***REMOVED*** */
				(this.watcher).pause();

				const fs = this.inputFileSystem;
				if (fs && fs.purge) {
					for (const item of changes) {
						fs.purge(item);
					***REMOVED***
					for (const item of removals) {
						fs.purge(item);
					***REMOVED***
				***REMOVED***
				const { fileTimeInfoEntries, contextTimeInfoEntries ***REMOVED*** = fetchTimeInfo();
				callback(
					null,
					fileTimeInfoEntries,
					contextTimeInfoEntries,
					changes,
					removals
				);
			***REMOVED***
		);

		this.watcher.watch({ files, directories, missing, startTime ***REMOVED***);

		if (oldWatcher) {
			oldWatcher.close();
		***REMOVED***
		return {
			close: () => {
				if (this.watcher) {
					this.watcher.close();
					this.watcher = null;
				***REMOVED***
			***REMOVED***,
			pause: () => {
				if (this.watcher) {
					this.watcher.pause();
				***REMOVED***
			***REMOVED***,
			getAggregatedRemovals: util.deprecate(
				() => {
					const items = this.watcher && this.watcher.aggregatedRemovals;
					const fs = this.inputFileSystem;
					if (items && fs && fs.purge) {
						for (const item of items) {
							fs.purge(item);
						***REMOVED***
					***REMOVED***
					return items;
				***REMOVED***,
				"Watcher.getAggregatedRemovals is deprecated in favor of Watcher.getInfo since that's more performant.",
				"DEP_WEBPACK_WATCHER_GET_AGGREGATED_REMOVALS"
			),
			getAggregatedChanges: util.deprecate(
				() => {
					const items = this.watcher && this.watcher.aggregatedChanges;
					const fs = this.inputFileSystem;
					if (items && fs && fs.purge) {
						for (const item of items) {
							fs.purge(item);
						***REMOVED***
					***REMOVED***
					return items;
				***REMOVED***,
				"Watcher.getAggregatedChanges is deprecated in favor of Watcher.getInfo since that's more performant.",
				"DEP_WEBPACK_WATCHER_GET_AGGREGATED_CHANGES"
			),
			getFileTimeInfoEntries: util.deprecate(
				() => fetchTimeInfo().fileTimeInfoEntries,
				"Watcher.getFileTimeInfoEntries is deprecated in favor of Watcher.getInfo since that's more performant.",
				"DEP_WEBPACK_WATCHER_FILE_TIME_INFO_ENTRIES"
			),
			getContextTimeInfoEntries: util.deprecate(
				() => fetchTimeInfo().contextTimeInfoEntries,
				"Watcher.getContextTimeInfoEntries is deprecated in favor of Watcher.getInfo since that's more performant.",
				"DEP_WEBPACK_WATCHER_CONTEXT_TIME_INFO_ENTRIES"
			),
			getInfo: () => {
				const removals = this.watcher && this.watcher.aggregatedRemovals;
				const changes = this.watcher && this.watcher.aggregatedChanges;
				const fs = this.inputFileSystem;
				if (fs && fs.purge) {
					if (removals) {
						for (const item of removals) {
							fs.purge(item);
						***REMOVED***
					***REMOVED***
					if (changes) {
						for (const item of changes) {
							fs.purge(item);
						***REMOVED***
					***REMOVED***
				***REMOVED***
				const { fileTimeInfoEntries, contextTimeInfoEntries ***REMOVED*** = fetchTimeInfo();
				return {
					changes,
					removals,
					fileTimeInfoEntries,
					contextTimeInfoEntries
				***REMOVED***;
			***REMOVED***
		***REMOVED***;
	***REMOVED***
***REMOVED***

module.exports = NodeWatchFileSystem;
