// webpack/lib/node/nodeConsole.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const util = require("util");
const truncateArgs = require("../logging/truncateArgs");

/** @typedef {import("../../declarations/WebpackOptions").InfrastructureLogging***REMOVED*** InfrastructureLogging */
/** @typedef {import("../logging/createConsoleLogger").LoggerConsole***REMOVED*** LoggerConsole */

/* eslint-disable no-console */

/**
 * @param {object***REMOVED*** options options
 * @param {boolean=***REMOVED*** options.colors colors
 * @param {boolean=***REMOVED*** options.appendOnly append only
 * @param {NonNullable<InfrastructureLogging["stream"]>***REMOVED*** options.stream stream
 * @returns {LoggerConsole***REMOVED*** logger function
 */
module.exports = ({ colors, appendOnly, stream ***REMOVED***) => {
	/** @type {string[] | undefined***REMOVED*** */
	let currentStatusMessage;
	let hasStatusMessage = false;
	let currentIndent = "";
	let currentCollapsed = 0;

	/**
	 * @param {string***REMOVED*** str string
	 * @param {string***REMOVED*** prefix prefix
	 * @param {string***REMOVED*** colorPrefix color prefix
	 * @param {string***REMOVED*** colorSuffix color suffix
	 * @returns {string***REMOVED*** indented string
	 */
	const indent = (str, prefix, colorPrefix, colorSuffix) => {
		if (str === "") return str;
		prefix = currentIndent + prefix;
		if (colors) {
			return (
				prefix +
				colorPrefix +
				str.replace(/\n/g, `${colorSuffix***REMOVED***\n${prefix***REMOVED***${colorPrefix***REMOVED***`) +
				colorSuffix
			);
		***REMOVED***

		return prefix + str.replace(/\n/g, `\n${prefix***REMOVED***`);
	***REMOVED***;

	const clearStatusMessage = () => {
		if (hasStatusMessage) {
			stream.write("\u001B[2K\r");
			hasStatusMessage = false;
		***REMOVED***
	***REMOVED***;

	const writeStatusMessage = () => {
		if (!currentStatusMessage) return;
		const l = stream.columns || 40;
		const args = truncateArgs(currentStatusMessage, l - 1);
		const str = args.join(" ");
		const coloredStr = `\u001B[1m${str***REMOVED***\u001B[39m\u001B[22m`;
		stream.write(`\u001B[2K\r${coloredStr***REMOVED***`);
		hasStatusMessage = true;
	***REMOVED***;

	/**
	 * @param {string***REMOVED*** prefix prefix
	 * @param {string***REMOVED*** colorPrefix color prefix
	 * @param {string***REMOVED*** colorSuffix color suffix
	 * @returns {(...args: EXPECTED_ANY[]) => void***REMOVED*** function to write with colors
	 */
	const writeColored =
		(prefix, colorPrefix, colorSuffix) =>
		(...args) => {
			if (currentCollapsed > 0) return;
			clearStatusMessage();
			const str = indent(
				util.format(...args),
				prefix,
				colorPrefix,
				colorSuffix
			);
			stream.write(`${str***REMOVED***\n`);
			writeStatusMessage();
		***REMOVED***;

	const writeGroupMessage = writeColored(
		"<-> ",
		"\u001B[1m\u001B[36m",
		"\u001B[39m\u001B[22m"
	);

	const writeGroupCollapsedMessage = writeColored(
		"<+> ",
		"\u001B[1m\u001B[36m",
		"\u001B[39m\u001B[22m"
	);

	return {
		log: writeColored("    ", "\u001B[1m", "\u001B[22m"),
		debug: writeColored("    ", "", ""),
		trace: writeColored("    ", "", ""),
		info: writeColored("<i> ", "\u001B[1m\u001B[32m", "\u001B[39m\u001B[22m"),
		warn: writeColored("<w> ", "\u001B[1m\u001B[33m", "\u001B[39m\u001B[22m"),
		error: writeColored("<e> ", "\u001B[1m\u001B[31m", "\u001B[39m\u001B[22m"),
		logTime: writeColored(
			"<t> ",
			"\u001B[1m\u001B[35m",
			"\u001B[39m\u001B[22m"
		),
		group: (...args) => {
			writeGroupMessage(...args);
			if (currentCollapsed > 0) {
				currentCollapsed++;
			***REMOVED*** else {
				currentIndent += "  ";
			***REMOVED***
		***REMOVED***,
		groupCollapsed: (...args) => {
			writeGroupCollapsedMessage(...args);
			currentCollapsed++;
		***REMOVED***,
		groupEnd: () => {
			if (currentCollapsed > 0) currentCollapsed--;
			else if (currentIndent.length >= 2)
				currentIndent = currentIndent.slice(0, -2);
		***REMOVED***,
		profile: console.profile && (name => console.profile(name)),
		profileEnd: console.profileEnd && (name => console.profileEnd(name)),
		clear:
			/** @type {() => void***REMOVED*** */
			(
				!appendOnly &&
					console.clear &&
					(() => {
						clearStatusMessage();
						console.clear();
						writeStatusMessage();
					***REMOVED***)
			),
		status: appendOnly
			? writeColored("<s> ", "", "")
			: (name, ...args) => {
					args = args.filter(Boolean);
					if (name === undefined && args.length === 0) {
						clearStatusMessage();
						currentStatusMessage = undefined;
					***REMOVED*** else if (
						typeof name === "string" &&
						name.startsWith("[webpack.Progress] ")
					) {
						currentStatusMessage = [name.slice(19), ...args];
						writeStatusMessage();
					***REMOVED*** else if (name === "[webpack.Progress]") {
						currentStatusMessage = [...args];
						writeStatusMessage();
					***REMOVED*** else {
						currentStatusMessage = [name, ...args];
						writeStatusMessage();
					***REMOVED***
				***REMOVED***
	***REMOVED***;
***REMOVED***;
