// webpack/lib/node/ReadFileCompileWasmPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const { WEBASSEMBLY_MODULE_TYPE_SYNC ***REMOVED*** = require("../ModuleTypeConstants");
const RuntimeGlobals = require("../RuntimeGlobals");
const Template = require("../Template");
const WasmChunkLoadingRuntimeModule = require("../wasm-sync/WasmChunkLoadingRuntimeModule");

/** @typedef {import("../Chunk")***REMOVED*** Chunk */
/** @typedef {import("../Compiler")***REMOVED*** Compiler */

/**
 * @typedef {object***REMOVED*** ReadFileCompileWasmPluginOptions
 * @property {boolean=***REMOVED*** mangleImports mangle imports
 * @property {boolean=***REMOVED*** import use import?
 */

// TODO webpack 6 remove

const PLUGIN_NAME = "ReadFileCompileWasmPlugin";

class ReadFileCompileWasmPlugin {
	/**
	 * @param {ReadFileCompileWasmPluginOptions=***REMOVED*** options options object
	 */
	constructor(options = {***REMOVED***) {
		this.options = options;
	***REMOVED***

	/**
	 * Apply the plugin
	 * @param {Compiler***REMOVED*** compiler the compiler instance
	 * @returns {void***REMOVED***
	 */
	apply(compiler) {
		compiler.hooks.thisCompilation.tap(PLUGIN_NAME, compilation => {
			const globalWasmLoading = compilation.outputOptions.wasmLoading;
			/**
			 * @param {Chunk***REMOVED*** chunk chunk
			 * @returns {boolean***REMOVED*** true, when wasm loading is enabled for the chunk
			 */
			const isEnabledForChunk = chunk => {
				const options = chunk.getEntryOptions();
				const wasmLoading =
					options && options.wasmLoading !== undefined
						? options.wasmLoading
						: globalWasmLoading;
				return wasmLoading === "async-node";
			***REMOVED***;

			/**
			 * @type {(path: string) => string***REMOVED*** callback to generate code to load the wasm file
			 */
			const generateLoadBinaryCode = this.options.import
				? path =>
						Template.asString([
							"Promise.all([import('fs'), import('url')]).then(([{ readFile ***REMOVED***, { URL ***REMOVED***]) => new Promise((resolve, reject) => {",
							Template.indent([
								`readFile(new URL(${path***REMOVED***, ${compilation.outputOptions.importMetaName***REMOVED***.url), (err, buffer) => {`,
								Template.indent([
									"if (err) return reject(err);",
									"",
									"// Fake fetch response",
									"resolve({",
									Template.indent(["arrayBuffer() { return buffer; ***REMOVED***"]),
									"***REMOVED***);"
								]),
								"***REMOVED***);"
							]),
							"***REMOVED***))"
						])
				: path =>
						Template.asString([
							"new Promise(function (resolve, reject) {",
							Template.indent([
								"var { readFile ***REMOVED*** = require('fs');",
								"var { join ***REMOVED*** = require('path');",
								"",
								"try {",
								Template.indent([
									`readFile(join(__dirname, ${path***REMOVED***), function(err, buffer){`,
									Template.indent([
										"if (err) return reject(err);",
										"",
										"// Fake fetch response",
										"resolve({",
										Template.indent(["arrayBuffer() { return buffer; ***REMOVED***"]),
										"***REMOVED***);"
									]),
									"***REMOVED***);"
								]),
								"***REMOVED*** catch (err) { reject(err); ***REMOVED***"
							]),
							"***REMOVED***)"
						]);

			compilation.hooks.runtimeRequirementInTree
				.for(RuntimeGlobals.ensureChunkHandlers)
				.tap(PLUGIN_NAME, (chunk, set, { chunkGraph ***REMOVED***) => {
					if (!isEnabledForChunk(chunk)) return;
					if (
						!chunkGraph.hasModuleInGraph(
							chunk,
							m => m.type === WEBASSEMBLY_MODULE_TYPE_SYNC
						)
					) {
						return;
					***REMOVED***
					set.add(RuntimeGlobals.moduleCache);
					compilation.addRuntimeModule(
						chunk,
						new WasmChunkLoadingRuntimeModule({
							generateLoadBinaryCode,
							supportsStreaming: false,
							mangleImports: this.options.mangleImports,
							runtimeRequirements: set
						***REMOVED***)
					);
				***REMOVED***);
		***REMOVED***);
	***REMOVED***
***REMOVED***

module.exports = ReadFileCompileWasmPlugin;
