// mini-css-extract-plugin/dist/hmr/hotModuleReplacement.js
"use strict";

/* eslint-env browser */
/*
  eslint-disable
  no-console,
  func-names
*/

/** @typedef {any***REMOVED*** TODO */

var normalizeUrl = require("./normalize-url");
var srcByModuleId = Object.create(null);
var noDocument = typeof document === "undefined";
var forEach = Array.prototype.forEach;

/**
 * @param {function***REMOVED*** fn
 * @param {number***REMOVED*** time
 * @returns {(function(): void)|****REMOVED***
 */
function debounce(fn, time) {
  var timeout = 0;
  return function () {
    // @ts-ignore
    var self = this;
    // eslint-disable-next-line prefer-rest-params
    var args = arguments;
    var functionCall = function functionCall() {
      return fn.apply(self, args);
    ***REMOVED***;
    clearTimeout(timeout);

    // @ts-ignore
    timeout = setTimeout(functionCall, time);
  ***REMOVED***;
***REMOVED***
function noop() {***REMOVED***

/**
 * @param {TODO***REMOVED*** moduleId
 * @returns {TODO***REMOVED***
 */
function getCurrentScriptUrl(moduleId) {
  var src = srcByModuleId[moduleId];
  if (!src) {
    if (document.currentScript) {
      src = ( /** @type {HTMLScriptElement***REMOVED*** */document.currentScript).src;
    ***REMOVED*** else {
      var scripts = document.getElementsByTagName("script");
      var lastScriptTag = scripts[scripts.length - 1];
      if (lastScriptTag) {
        src = lastScriptTag.src;
      ***REMOVED***
    ***REMOVED***
    srcByModuleId[moduleId] = src;
  ***REMOVED***

  /**
   * @param {string***REMOVED*** fileMap
   * @returns {null | string[]***REMOVED***
   */
  return function (fileMap) {
    if (!src) {
      return null;
    ***REMOVED***
    var splitResult = src.split(/([^\\/]+)\.js$/);
    var filename = splitResult && splitResult[1];
    if (!filename) {
      return [src.replace(".js", ".css")];
    ***REMOVED***
    if (!fileMap) {
      return [src.replace(".js", ".css")];
    ***REMOVED***
    return fileMap.split(",").map(function (mapRule) {
      var reg = new RegExp("".concat(filename, "\\.js$"), "g");
      return normalizeUrl(src.replace(reg, "".concat(mapRule.replace(/{fileName***REMOVED***/g, filename), ".css")));
    ***REMOVED***);
  ***REMOVED***;
***REMOVED***

/**
 * @param {TODO***REMOVED*** el
 * @param {string***REMOVED*** [url]
 */
function updateCss(el, url) {
  if (!url) {
    if (!el.href) {
      return;
    ***REMOVED***

    // eslint-disable-next-line
    url = el.href.split("?")[0];
  ***REMOVED***
  if (!isUrlRequest( /** @type {string***REMOVED*** */url)) {
    return;
  ***REMOVED***
  if (el.isLoaded === false) {
    // We seem to be about to replace a css link that hasn't loaded yet.
    // We're probably changing the same file more than once.
    return;
  ***REMOVED***
  if (!url || !(url.indexOf(".css") > -1)) {
    return;
  ***REMOVED***

  // eslint-disable-next-line no-param-reassign
  el.visited = true;
  var newEl = el.cloneNode();
  newEl.isLoaded = false;
  newEl.addEventListener("load", function () {
    if (newEl.isLoaded) {
      return;
    ***REMOVED***
    newEl.isLoaded = true;
    el.parentNode.removeChild(el);
  ***REMOVED***);
  newEl.addEventListener("error", function () {
    if (newEl.isLoaded) {
      return;
    ***REMOVED***
    newEl.isLoaded = true;
    el.parentNode.removeChild(el);
  ***REMOVED***);
  newEl.href = "".concat(url, "?").concat(Date.now());
  if (el.nextSibling) {
    el.parentNode.insertBefore(newEl, el.nextSibling);
  ***REMOVED*** else {
    el.parentNode.appendChild(newEl);
  ***REMOVED***
***REMOVED***

/**
 * @param {string***REMOVED*** href
 * @param {TODO***REMOVED*** src
 * @returns {TODO***REMOVED***
 */
function getReloadUrl(href, src) {
  var ret;

  // eslint-disable-next-line no-param-reassign
  href = normalizeUrl(href);
  src.some(
  /**
   * @param {string***REMOVED*** url
   */
  // eslint-disable-next-line array-callback-return
  function (url) {
    if (href.indexOf(src) > -1) {
      ret = url;
    ***REMOVED***
  ***REMOVED***);
  return ret;
***REMOVED***

/**
 * @param {string***REMOVED*** [src]
 * @returns {boolean***REMOVED***
 */
function reloadStyle(src) {
  if (!src) {
    return false;
  ***REMOVED***
  var elements = document.querySelectorAll("link");
  var loaded = false;
  forEach.call(elements, function (el) {
    if (!el.href) {
      return;
    ***REMOVED***
    var url = getReloadUrl(el.href, src);
    if (!isUrlRequest(url)) {
      return;
    ***REMOVED***
    if (el.visited === true) {
      return;
    ***REMOVED***
    if (url) {
      updateCss(el, url);
      loaded = true;
    ***REMOVED***
  ***REMOVED***);
  return loaded;
***REMOVED***
function reloadAll() {
  var elements = document.querySelectorAll("link");
  forEach.call(elements, function (el) {
    if (el.visited === true) {
      return;
    ***REMOVED***
    updateCss(el);
  ***REMOVED***);
***REMOVED***

/**
 * @param {string***REMOVED*** url
 * @returns {boolean***REMOVED***
 */
function isUrlRequest(url) {
  // An URL is not an request if

  // It is not http or https
  if (!/^[a-zA-Z][a-zA-Z\d+\-.]*:/.test(url)) {
    return false;
  ***REMOVED***
  return true;
***REMOVED***

/**
 * @param {TODO***REMOVED*** moduleId
 * @param {TODO***REMOVED*** options
 * @returns {TODO***REMOVED***
 */
module.exports = function (moduleId, options) {
  if (noDocument) {
    console.log("no window.document found, will not HMR CSS");
    return noop;
  ***REMOVED***
  var getScriptSrc = getCurrentScriptUrl(moduleId);
  function update() {
    var src = getScriptSrc(options.filename);
    var reloaded = reloadStyle(src);
    if (options.locals) {
      console.log("[HMR] Detected local css modules. Reload all css");
      reloadAll();
      return;
    ***REMOVED***
    if (reloaded) {
      console.log("[HMR] css reload %s", src.join(" "));
    ***REMOVED*** else {
      console.log("[HMR] Reload all css");
      reloadAll();
    ***REMOVED***
  ***REMOVED***
  return debounce(update, 50);
***REMOVED***;
