// tailwindcss/lib/lib/load-config.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
***REMOVED***);
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    ***REMOVED***);
***REMOVED***
_export(exports, {
    useCustomJiti: function() {
        return useCustomJiti;
    ***REMOVED***,
    loadConfig: function() {
        return loadConfig;
    ***REMOVED***
***REMOVED***);
const _jiti = /*#__PURE__*/ _interop_require_default(require("jiti"));
const _sucrase = require("sucrase");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    ***REMOVED***;
***REMOVED***
let jiti = null;
function useCustomJiti(_jiti) {
    jiti = _jiti();
***REMOVED***
function lazyJiti() {
    return jiti !== null && jiti !== void 0 ? jiti : jiti = (0, _jiti.default)(__filename, {
        interopDefault: true,
        transform: (opts)=>{
            // Sucrase can't transform import.meta so we have to use Babel
            if (opts.source.includes("import.meta")) {
                return require("jiti/dist/babel.js")(opts);
            ***REMOVED***
            return (0, _sucrase.transform)(opts.source, {
                transforms: [
                    "typescript",
                    "imports"
                ]
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***);
***REMOVED***
function loadConfig(path) {
    let config = function() {
        if (!path) return {***REMOVED***;
        // Always use jiti for now. There is a a bug that occurs in Node v22.12+
        // where imported files return invalid results
        return lazyJiti()(path);
        // Always use jiti for ESM or TS files
        if (path && (path.endsWith(".mjs") || path.endsWith(".ts") || path.endsWith(".cts") || path.endsWith(".mts"))) {
            return lazyJiti()(path);
        ***REMOVED***
        try {
            return path ? require(path) : {***REMOVED***;
        ***REMOVED*** catch  {
            return lazyJiti()(path);
        ***REMOVED***
    ***REMOVED***();
    var _config_default;
    return (_config_default = config.default) !== null && _config_default !== void 0 ? _config_default : config;
***REMOVED***
