// tailwindcss/lib/lib/regex.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
***REMOVED***);
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    ***REMOVED***);
***REMOVED***
_export(exports, {
    pattern: function() {
        return pattern;
    ***REMOVED***,
    withoutCapturing: function() {
        return withoutCapturing;
    ***REMOVED***,
    any: function() {
        return any;
    ***REMOVED***,
    optional: function() {
        return optional;
    ***REMOVED***,
    zeroOrMore: function() {
        return zeroOrMore;
    ***REMOVED***,
    nestedBrackets: function() {
        return nestedBrackets;
    ***REMOVED***,
    escape: function() {
        return escape;
    ***REMOVED***
***REMOVED***);
const REGEX_SPECIAL = /[\\^$.*+?()[\]{***REMOVED***|]/g;
const REGEX_HAS_SPECIAL = RegExp(REGEX_SPECIAL.source);
/**
 * @param {string|RegExp|Array<string|RegExp>***REMOVED*** source
 */ function toSource(source) {
    source = Array.isArray(source) ? source : [
        source
    ];
    source = source.map((item)=>item instanceof RegExp ? item.source : item);
    return source.join("");
***REMOVED***
function pattern(source) {
    return new RegExp(toSource(source), "g");
***REMOVED***
function withoutCapturing(source) {
    return new RegExp(`(?:${toSource(source)***REMOVED***)`, "g");
***REMOVED***
function any(sources) {
    return `(?:${sources.map(toSource).join("|")***REMOVED***)`;
***REMOVED***
function optional(source) {
    return `(?:${toSource(source)***REMOVED***)?`;
***REMOVED***
function zeroOrMore(source) {
    return `(?:${toSource(source)***REMOVED***)*`;
***REMOVED***
function nestedBrackets(open, close, depth = 1) {
    return withoutCapturing([
        escape(open),
        /[^\s]*/,
        depth === 1 ? `[^${escape(open)***REMOVED***${escape(close)***REMOVED***\s]*` : any([
            `[^${escape(open)***REMOVED***${escape(close)***REMOVED***\s]*`,
            nestedBrackets(open, close, depth - 1)
        ]),
        /[^\s]*/,
        escape(close)
    ]);
***REMOVED***
function escape(string) {
    return string && REGEX_HAS_SPECIAL.test(string) ? string.replace(REGEX_SPECIAL, "\\$&") : string || "";
***REMOVED***
