// tailwindcss/lib/lib/sharedState.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
***REMOVED***);
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    ***REMOVED***);
***REMOVED***
_export(exports, {
    env: function() {
        return env;
    ***REMOVED***,
    contextMap: function() {
        return contextMap;
    ***REMOVED***,
    configContextMap: function() {
        return configContextMap;
    ***REMOVED***,
    contextSourcesMap: function() {
        return contextSourcesMap;
    ***REMOVED***,
    sourceHashMap: function() {
        return sourceHashMap;
    ***REMOVED***,
    NOT_ON_DEMAND: function() {
        return NOT_ON_DEMAND;
    ***REMOVED***,
    NONE: function() {
        return NONE;
    ***REMOVED***,
    resolveDebug: function() {
        return resolveDebug;
    ***REMOVED***
***REMOVED***);
const env = typeof process !== "undefined" ? {
    NODE_ENV: process.env.NODE_ENV,
    DEBUG: resolveDebug(process.env.DEBUG)
***REMOVED*** : {
    NODE_ENV: "production",
    DEBUG: false
***REMOVED***;
const contextMap = new Map();
const configContextMap = new Map();
const contextSourcesMap = new Map();
const sourceHashMap = new Map();
const NOT_ON_DEMAND = new String("*");
const NONE = Symbol("__NONE__");
function resolveDebug(debug) {
    if (debug === undefined) {
        return false;
    ***REMOVED***
    // Environment variables are strings, so convert to boolean
    if (debug === "true" || debug === "1") {
        return true;
    ***REMOVED***
    if (debug === "false" || debug === "0") {
        return false;
    ***REMOVED***
    // Keep the debug convention into account:
    // DEBUG=* -> This enables all debug modes
    // DEBUG=projectA,projectB,projectC -> This enables debug for projectA, projectB and projectC
    // DEBUG=projectA:* -> This enables all debug modes for projectA (if you have sub-types)
    // DEBUG=projectA,-projectB -> This enables debug for projectA and explicitly disables it for projectB
    if (debug === "*") {
        return true;
    ***REMOVED***
    let debuggers = debug.split(",").map((d)=>d.split(":")[0]);
    // Ignoring tailwindcss
    if (debuggers.includes("-tailwindcss")) {
        return false;
    ***REMOVED***
    // Including tailwindcss
    if (debuggers.includes("tailwindcss")) {
        return true;
    ***REMOVED***
    return false;
***REMOVED***
