// renderkid/lib/renderKid/Styles.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); ***REMOVED*** ***REMOVED***

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); ***REMOVED*** ***REMOVED***

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; ***REMOVED***

// Generated by CoffeeScript 2.5.1
var MixedDeclarationSet, StyleSheet, Styles, terminalWidth;
StyleSheet = require('./styles/StyleSheet');
MixedDeclarationSet = require('./styles/rule/MixedDeclarationSet');
terminalWidth = require('../tools').getCols();

module.exports = Styles = function () {
  var self;

  var Styles = /*#__PURE__*/function () {
    function Styles() {
      _classCallCheck(this, Styles);

      this._defaultStyles = new StyleSheet();
      this._userStyles = new StyleSheet();

      this._setDefaultStyles();
    ***REMOVED***

    _createClass(Styles, [{
      key: "_setDefaultStyles",
      value: function _setDefaultStyles() {
        this._defaultStyles.setRule(self.defaultRules);
      ***REMOVED***
    ***REMOVED***, {
      key: "setRule",
      value: function setRule(selector, rules) {
        this._userStyles.setRule.apply(this._userStyles, arguments);

        return this;
      ***REMOVED***
    ***REMOVED***, {
      key: "getStyleFor",
      value: function getStyleFor(el) {
        var styles;
        styles = el.styles;

        if (styles == null) {
          el.styles = styles = this._getComputedStyleFor(el);
        ***REMOVED***

        return styles;
      ***REMOVED***
    ***REMOVED***, {
      key: "_getRawStyleFor",
      value: function _getRawStyleFor(el) {
        var def, user;
        def = this._defaultStyles.getRulesFor(el);
        user = this._userStyles.getRulesFor(el);
        return MixedDeclarationSet.mix(def, user).toObject();
      ***REMOVED***
    ***REMOVED***, {
      key: "_getComputedStyleFor",
      value: function _getComputedStyleFor(el) {
        var decs, parent, prop, ref, val;
        decs = {***REMOVED***;
        parent = el.parent;
        ref = this._getRawStyleFor(el);

        for (prop in ref) {
          val = ref[prop];

          if (val !== 'inherit') {
            decs[prop] = val;
          ***REMOVED*** else {
            throw Error("Inherited styles are not supported yet.");
          ***REMOVED***
        ***REMOVED***

        return decs;
      ***REMOVED***
    ***REMOVED***]);

    return Styles;
  ***REMOVED***();

  ;
  self = Styles;
  Styles.defaultRules = {
    '*': {
      display: 'inline'
    ***REMOVED***,
    'body': {
      background: 'none',
      color: 'white',
      display: 'block',
      width: terminalWidth + ' !important'
    ***REMOVED***
  ***REMOVED***;
  return Styles;
***REMOVED***.call(void 0);
