// @babel/helper-define-polyfill-provider/lib/visitors/usage.js
"use strict";

exports.__esModule = true;
exports.default = void 0;
var _utils = require("../utils");
function isRemoved(path) {
  if (path.removed) return true;
  if (!path.parentPath) return false;
  if (path.listKey) {
    var _path$parentPath$node;
    if (!((_path$parentPath$node = path.parentPath.node) != null && (_path$parentPath$node = _path$parentPath$node[path.listKey]) != null && _path$parentPath$node.includes(path.node))) return true;
  ***REMOVED*** else {
    if (path.parentPath.node[path.key] !== path.node) return true;
  ***REMOVED***
  return isRemoved(path.parentPath);
***REMOVED***
var _default = callProvider => {
  function property(object, key, placement, path) {
    return callProvider({
      kind: "property",
      object,
      key,
      placement
    ***REMOVED***, path);
  ***REMOVED***
  function handleReferencedIdentifier(path) {
    const {
      node: {
        name
      ***REMOVED***,
      scope
    ***REMOVED*** = path;
    if (scope.getBindingIdentifier(name)) return;
    callProvider({
      kind: "global",
      name
    ***REMOVED***, path);
  ***REMOVED***
  function analyzeMemberExpression(path) {
    const key = (0, _utils.resolveKey)(path.get("property"), path.node.computed);
    return {
      key,
      handleAsMemberExpression: !!key && key !== "prototype"
    ***REMOVED***;
  ***REMOVED***
  return {
    // Symbol(), new Promise
    ReferencedIdentifier(path) {
      const {
        parentPath
      ***REMOVED*** = path;
      if (parentPath.isMemberExpression({
        object: path.node
      ***REMOVED***) && analyzeMemberExpression(parentPath).handleAsMemberExpression) {
        return;
      ***REMOVED***
      handleReferencedIdentifier(path);
    ***REMOVED***,
    "MemberExpression|OptionalMemberExpression"(path) {
      const {
        key,
        handleAsMemberExpression
      ***REMOVED*** = analyzeMemberExpression(path);
      if (!handleAsMemberExpression) return;
      const object = path.get("object");
      let objectIsGlobalIdentifier = object.isIdentifier();
      if (objectIsGlobalIdentifier) {
        const binding = object.scope.getBinding(object.node.name);
        if (binding) {
          if (binding.path.isImportNamespaceSpecifier()) return;
          objectIsGlobalIdentifier = false;
        ***REMOVED***
      ***REMOVED***
      const source = (0, _utils.resolveSource)(object);
      let skipObject = property(source.id, key, source.placement, path);
      skipObject || (skipObject = !objectIsGlobalIdentifier || path.shouldSkip || object.shouldSkip || isRemoved(object));
      if (!skipObject) handleReferencedIdentifier(object);
    ***REMOVED***,
    ObjectPattern(path) {
      const {
        parentPath,
        parent
      ***REMOVED*** = path;
      let obj;

      // const { keys, values ***REMOVED*** = Object
      if (parentPath.isVariableDeclarator()) {
        obj = parentPath.get("init");
        // ({ keys, values ***REMOVED*** = Object)
      ***REMOVED*** else if (parentPath.isAssignmentExpression()) {
        obj = parentPath.get("right");
        // !function ({ keys, values ***REMOVED***) {...***REMOVED*** (Object)
        // resolution does not work after properties transform :-(
      ***REMOVED*** else if (parentPath.isFunction()) {
        const grand = parentPath.parentPath;
        if (grand.isCallExpression() || grand.isNewExpression()) {
          if (grand.node.callee === parent) {
            obj = grand.get("arguments")[path.key];
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
      let id = null;
      let placement = null;
      if (obj) ({
        id,
        placement
      ***REMOVED*** = (0, _utils.resolveSource)(obj));
      for (const prop of path.get("properties")) {
        if (prop.isObjectProperty()) {
          const key = (0, _utils.resolveKey)(prop.get("key"));
          if (key) property(id, key, placement, prop);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***,
    BinaryExpression(path) {
      if (path.node.operator !== "in") return;
      const source = (0, _utils.resolveSource)(path.get("right"));
      const key = (0, _utils.resolveKey)(path.get("left"), true);
      if (!key) return;
      callProvider({
        kind: "in",
        object: source.id,
        key,
        placement: source.placement
      ***REMOVED***, path);
    ***REMOVED***
  ***REMOVED***;
***REMOVED***;
exports.default = _default;
