// jsdom/lib/jsdom/living/window/Location-impl.js
"use strict";
const whatwgURL = require("whatwg-url");
const DOMException = require("domexception/webidl2js-wrapper");
const { documentBaseURL, parseURLToResultingURLRecord ***REMOVED*** = require("../helpers/document-base-url");
const { navigate ***REMOVED*** = require("./navigation");

// Not implemented: use of entry settings object's API base URL in href setter, assign, and replace. Instead we just
// use the document base URL. The difference matters in the case of cross-frame calls.

exports.implementation = class LocationImpl {
  constructor(globalObject, args, privateData) {
    this._relevantDocument = privateData.relevantDocument;
    this.url = null;

    this._globalObject = globalObject;
  ***REMOVED***

  get _url() {
    return this._relevantDocument._URL;
  ***REMOVED***

  _locationObjectSetterNavigate(url) {
    // Not implemented: extra steps here to determine replacement flag.

    return this._locationObjectNavigate(url);
  ***REMOVED***

  _locationObjectNavigate(url, { replacement = false ***REMOVED*** = {***REMOVED***) {
    // Not implemented: the setup for calling navigate, which doesn't apply to our stub navigate anyway.

    navigate(this._relevantDocument._defaultView, url, { replacement, exceptionsEnabled: true ***REMOVED***);
  ***REMOVED***

  toString() {
    return this.href;
  ***REMOVED***

  get href() {
    return whatwgURL.serializeURL(this._url);
  ***REMOVED***
  set href(v) {
    const newURL = whatwgURL.parseURL(v, { baseURL: documentBaseURL(this._relevantDocument) ***REMOVED***);
    if (newURL === null) {
      throw new TypeError(`Could not parse "${v***REMOVED***" as a URL`);
    ***REMOVED***

    this._locationObjectSetterNavigate(newURL);
  ***REMOVED***

  get origin() {
    return whatwgURL.serializeURLOrigin(this._url);
  ***REMOVED***

  get protocol() {
    return this._url.scheme + ":";
  ***REMOVED***
  set protocol(v) {
    const copyURL = { ...this._url ***REMOVED***;

    const possibleFailure = whatwgURL.basicURLParse(v + ":", { url: copyURL, stateOverride: "scheme start" ***REMOVED***);
    if (possibleFailure === null) {
      throw new TypeError(`Could not parse the URL after setting the procol to "${v***REMOVED***"`);
    ***REMOVED***

    if (copyURL.scheme !== "http" && copyURL.scheme !== "https") {
      return;
    ***REMOVED***

    this._locationObjectSetterNavigate(copyURL);
  ***REMOVED***

  get host() {
    const url = this._url;

    if (url.host === null) {
      return "";
    ***REMOVED***
    if (url.port === null) {
      return whatwgURL.serializeHost(url.host);
    ***REMOVED***

    return whatwgURL.serializeHost(url.host) + ":" + whatwgURL.serializeInteger(url.port);
  ***REMOVED***
  set host(v) {
    const copyURL = { ...this._url ***REMOVED***;

    if (copyURL.cannotBeABaseURL) {
      return;
    ***REMOVED***

    whatwgURL.basicURLParse(v, { url: copyURL, stateOverride: "host" ***REMOVED***);

    this._locationObjectSetterNavigate(copyURL);
  ***REMOVED***

  get hostname() {
    if (this._url.host === null) {
      return "";
    ***REMOVED***

    return whatwgURL.serializeHost(this._url.host);
  ***REMOVED***
  set hostname(v) {
    const copyURL = { ...this._url ***REMOVED***;

    if (copyURL.cannotBeABaseURL) {
      return;
    ***REMOVED***

    whatwgURL.basicURLParse(v, { url: copyURL, stateOverride: "hostname" ***REMOVED***);

    this._locationObjectSetterNavigate(copyURL);
  ***REMOVED***

  get port() {
    if (this._url.port === null) {
      return "";
    ***REMOVED***

    return whatwgURL.serializeInteger(this._url.port);
  ***REMOVED***
  set port(v) {
    const copyURL = { ...this._url ***REMOVED***;

    if (copyURL.host === null || copyURL.cannotBeABaseURL || copyURL.scheme === "file") {
      return;
    ***REMOVED***

    whatwgURL.basicURLParse(v, { url: copyURL, stateOverride: "port" ***REMOVED***);

    this._locationObjectSetterNavigate(copyURL);
  ***REMOVED***

  get pathname() {
    const url = this._url;

    if (url.cannotBeABaseURL) {
      return url.path[0];
    ***REMOVED***

    return "/" + url.path.join("/");
  ***REMOVED***
  set pathname(v) {
    const copyURL = { ...this._url ***REMOVED***;

    if (copyURL.cannotBeABaseURL) {
      return;
    ***REMOVED***

    copyURL.path = [];
    whatwgURL.basicURLParse(v, { url: copyURL, stateOverride: "path start" ***REMOVED***);

    this._locationObjectSetterNavigate(copyURL);
  ***REMOVED***

  get search() {
    if (this._url.query === null || this._url.query === "") {
      return "";
    ***REMOVED***

    return "?" + this._url.query;
  ***REMOVED***
  set search(v) {
    const copyURL = { ...this._url ***REMOVED***;

    if (v === "") {
      copyURL.query = null;
    ***REMOVED*** else {
      const input = v[0] === "?" ? v.substring(1) : v;
      copyURL.query = "";
      whatwgURL.basicURLParse(input, {
        url: copyURL,
        stateOverride: "query",
        encodingOverride: this._relevantDocument.charset
      ***REMOVED***);
    ***REMOVED***

    this._locationObjectSetterNavigate(copyURL);
  ***REMOVED***

  get hash() {
    if (this._url.fragment === null || this._url.fragment === "") {
      return "";
    ***REMOVED***

    return "#" + this._url.fragment;
  ***REMOVED***
  set hash(v) {
    const copyURL = { ...this._url ***REMOVED***;

    if (copyURL.scheme === "javascript") {
      return;
    ***REMOVED***

    if (v === "") {
      copyURL.fragment = null;
    ***REMOVED*** else {
      const input = v[0] === "#" ? v.substring(1) : v;
      copyURL.fragment = "";
      whatwgURL.basicURLParse(input, { url: copyURL, stateOverride: "fragment" ***REMOVED***);
    ***REMOVED***

    this._locationObjectSetterNavigate(copyURL);
  ***REMOVED***

  assign(url) {
    // Should be entry settings object; oh well
    const parsedURL = parseURLToResultingURLRecord(url, this._relevantDocument);

    if (parsedURL === null) {
      throw DOMException.create(this._globalObject, [
        `Could not resolve the given string "${url***REMOVED***" relative to the base URL "${this._relevantDocument.URL***REMOVED***"`,
        "SyntaxError"
      ]);
    ***REMOVED***

    this._locationObjectNavigate(parsedURL);
  ***REMOVED***

  replace(url) {
    // Should be entry settings object; oh well
    const parsedURL = parseURLToResultingURLRecord(url, this._relevantDocument);

    if (parsedURL === null) {
      throw DOMException.create(this._globalObject, [
        `Could not resolve the given string "${url***REMOVED***" relative to the base URL "${this._relevantDocument.URL***REMOVED***"`,
        "SyntaxError"
      ]);
    ***REMOVED***

    this._locationObjectNavigate(parsedURL, { replacement: true ***REMOVED***);
  ***REMOVED***

  reload() {
    const flags = { replace: true, reloadTriggered: true, exceptionsEnabled: true ***REMOVED***;
    navigate(this._relevantDocument._defaultView, this._url, flags);
  ***REMOVED***
***REMOVED***;
