// csso/lib/restructure/3-disjoinRuleset.js
var List = require('css-tree').List;
var walk = require('css-tree').walk;

function processRule(node, item, list) {
    var selectors = node.prelude.children;

    // generate new rule sets:
    // .a, .b { color: red; ***REMOVED***
    // ->
    // .a { color: red; ***REMOVED***
    // .b { color: red; ***REMOVED***

    // while there are more than 1 simple selector split for rulesets
    while (selectors.head !== selectors.tail) {
        var newSelectors = new List();
        newSelectors.insert(selectors.remove(selectors.head));

        list.insert(list.createItem({
            type: 'Rule',
            loc: node.loc,
            prelude: {
                type: 'SelectorList',
                loc: node.prelude.loc,
                children: newSelectors
            ***REMOVED***,
            block: {
                type: 'Block',
                loc: node.block.loc,
                children: node.block.children.copy()
            ***REMOVED***,
            pseudoSignature: node.pseudoSignature
        ***REMOVED***), item);
    ***REMOVED***
***REMOVED***

module.exports = function disjoinRule(ast) {
    walk(ast, {
        visit: 'Rule',
        reverse: true,
        enter: processRule
    ***REMOVED***);
***REMOVED***;
