// schema-utils/dist/keywords/undefinedAsNull.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

/** @typedef {import("ajv").Ajv***REMOVED*** Ajv */

/**
 *
 * @param {Ajv***REMOVED*** ajv
 * @param {string***REMOVED*** keyword
 * @param {any***REMOVED*** definition
 */
function addKeyword(ajv, keyword, definition) {
  let customRuleCode;

  try {
    // @ts-ignore
    // eslint-disable-next-line global-require
    customRuleCode = require("ajv/lib/dotjs/custom"); // @ts-ignore

    const {
      RULES
    ***REMOVED*** = ajv;
    let ruleGroup;

    for (let i = 0; i < RULES.length; i++) {
      const rg = RULES[i];

      if (typeof rg.type === "undefined") {
        ruleGroup = rg;
        break;
      ***REMOVED***
    ***REMOVED***

    const rule = {
      keyword,
      definition,
      custom: true,
      code: customRuleCode,
      implements: definition.implements
    ***REMOVED***;
    ruleGroup.rules.unshift(rule);
    RULES.custom[keyword] = rule;
    RULES.keywords[keyword] = true;
    RULES.all[keyword] = true;
  ***REMOVED*** catch (e) {// Nothing, fallback
  ***REMOVED***
***REMOVED***
/**
 *
 * @param {Ajv***REMOVED*** ajv
 * @returns {Ajv***REMOVED***
 */


function addUndefinedAsNullKeyword(ajv) {
  // There is workaround for old versions of ajv, where `before` is not implemented
  addKeyword(ajv, "undefinedAsNull", {
    modifying: true,

    /**
     * @param {boolean***REMOVED*** kwVal
     * @param {unknown***REMOVED*** data
     * @param {any***REMOVED*** parentSchema
     * @param {string***REMOVED*** dataPath
     * @param {unknown***REMOVED*** parentData
     * @param {number | string***REMOVED*** parentDataProperty
     * @return {boolean***REMOVED***
     */
    validate(kwVal, data, parentSchema, dataPath, parentData, parentDataProperty) {
      if (kwVal && parentSchema && typeof parentSchema.enum !== "undefined" && parentData && typeof parentDataProperty === "number") {
        const idx =
        /** @type {number***REMOVED*** */
        parentDataProperty;
        const parentDataRef =
        /** @type {any[]***REMOVED*** */
        parentData;

        if (typeof parentDataRef[idx] === "undefined") {
          parentDataRef[idx] = null;
        ***REMOVED***
      ***REMOVED***

      return true;
    ***REMOVED***

  ***REMOVED***);
  return ajv;
***REMOVED***

var _default = addUndefinedAsNullKeyword;
exports.default = _default;
