// source-map-loader/dist/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.fetchFromURL = fetchFromURL;
exports.flattenSourceMap = flattenSourceMap;
exports.getSourceMappingURL = getSourceMappingURL;
exports.isURL = isURL;

var _path = _interopRequireDefault(require("path"));

var _url = _interopRequireDefault(require("url"));

var _sourceMapJs = _interopRequireDefault(require("source-map-js"));

var _iconvLite = require("iconv-lite");

var _parseDataUrl = _interopRequireDefault(require("./parse-data-url"));

var _labelsToNames = _interopRequireDefault(require("./labels-to-names"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj ***REMOVED***; ***REMOVED***

// Matches only the last occurrence of sourceMappingURL
const innerRegex = /\s*[#@]\s*sourceMappingURL\s*=\s*([^\s'"]*)\s*/;
/* eslint-disable prefer-template */

const sourceMappingURLRegex = RegExp("(?:" + "/\\*" + "(?:\\s*\r?\n(?://)?)?" + "(?:" + innerRegex.source + ")" + "\\s*" + "\\*/" + "|" + "//(?:" + innerRegex.source + ")" + ")" + "\\s*");
/* eslint-enable prefer-template */

function labelToName(label) {
  const labelLowercase = String(label).trim().toLowerCase();
  return _labelsToNames.default[labelLowercase] || null;
***REMOVED***

async function flattenSourceMap(map) {
  const consumer = await new _sourceMapJs.default.SourceMapConsumer(map);
  const generatedMap = map.file ? new _sourceMapJs.default.SourceMapGenerator({
    file: map.file
  ***REMOVED***) : new _sourceMapJs.default.SourceMapGenerator();
  consumer.sources.forEach(sourceFile => {
    const sourceContent = consumer.sourceContentFor(sourceFile, true);
    generatedMap.setSourceContent(sourceFile, sourceContent);
  ***REMOVED***);
  consumer.eachMapping(mapping => {
    const {
      source
    ***REMOVED*** = consumer.originalPositionFor({
      line: mapping.generatedLine,
      column: mapping.generatedColumn
    ***REMOVED***);
    const mappings = {
      source,
      original: {
        line: mapping.originalLine,
        column: mapping.originalColumn
      ***REMOVED***,
      generated: {
        line: mapping.generatedLine,
        column: mapping.generatedColumn
      ***REMOVED***
    ***REMOVED***;

    if (source) {
      generatedMap.addMapping(mappings);
    ***REMOVED***
  ***REMOVED***);
  return generatedMap.toJSON();
***REMOVED***

function getSourceMappingURL(code) {
  const lines = code.split(/^/m);
  let match;

  for (let i = lines.length - 1; i >= 0; i--) {
    match = lines[i].match(sourceMappingURLRegex);

    if (match) {
      break;
    ***REMOVED***
  ***REMOVED***

  const sourceMappingURL = match ? match[1] || match[2] || "" : null;
  return {
    sourceMappingURL: sourceMappingURL ? decodeURI(sourceMappingURL) : sourceMappingURL,
    replacementString: match ? match[0] : null
  ***REMOVED***;
***REMOVED***

function getAbsolutePath(context, request, sourceRoot) {
  if (isURL(sourceRoot)) {
    return new URL(request, sourceRoot).toString();
  ***REMOVED***

  if (sourceRoot) {
    if (_path.default.isAbsolute(sourceRoot)) {
      return _path.default.join(sourceRoot, request);
    ***REMOVED***

    return _path.default.join(context, sourceRoot, request);
  ***REMOVED***

  return _path.default.join(context, request);
***REMOVED***

function fetchFromDataURL(loaderContext, sourceURL) {
  const dataURL = (0, _parseDataUrl.default)(sourceURL);

  if (dataURL) {
    // https://tools.ietf.org/html/rfc4627
    // JSON text SHALL be encoded in Unicode. The default encoding is UTF-8.
    const encodingName = labelToName(dataURL.parameters.get("charset")) || "UTF-8";
    return (0, _iconvLite.decode)(dataURL.body, encodingName);
  ***REMOVED***

  throw new Error(`Failed to parse source map from "data" URL: ${sourceURL***REMOVED***`);
***REMOVED***

async function fetchFromFilesystem(loaderContext, sourceURL) {
  let buffer;

  if (isURL(sourceURL)) {
    return {
      path: sourceURL
    ***REMOVED***;
  ***REMOVED***

  try {
    buffer = await new Promise((resolve, reject) => {
      loaderContext.fs.readFile(sourceURL, (error, data) => {
        if (error) {
          return reject(error);
        ***REMOVED***

        return resolve(data);
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED*** catch (error) {
    throw new Error(`Failed to parse source map from '${sourceURL***REMOVED***' file: ${error***REMOVED***`);
  ***REMOVED***

  return {
    path: sourceURL,
    data: buffer.toString()
  ***REMOVED***;
***REMOVED***

async function fetchPathsFromFilesystem(loaderContext, possibleRequests, errorsAccumulator = "") {
  let result;

  try {
    result = await fetchFromFilesystem(loaderContext, possibleRequests[0], errorsAccumulator);
  ***REMOVED*** catch (error) {
    // eslint-disable-next-line no-param-reassign
    errorsAccumulator += `${error.message***REMOVED***\n\n`;
    const [, ...tailPossibleRequests] = possibleRequests;

    if (tailPossibleRequests.length === 0) {
      error.message = errorsAccumulator;
      throw error;
    ***REMOVED***

    return fetchPathsFromFilesystem(loaderContext, tailPossibleRequests, errorsAccumulator);
  ***REMOVED***

  return result;
***REMOVED***

function isURL(value) {
  return /^[a-z][a-z0-9+.-]*:/i.test(value) && !_path.default.win32.isAbsolute(value);
***REMOVED***

async function fetchFromURL(loaderContext, context, url, sourceRoot, skipReading = false) {
  // 1. It's an absolute url and it is not `windows` path like `C:\dir\file`
  if (isURL(url)) {
    const {
      protocol
    ***REMOVED*** = _url.default.parse(url);

    if (protocol === "data:") {
      if (skipReading) {
        return {
          sourceURL: ""
        ***REMOVED***;
      ***REMOVED***

      const sourceContent = fetchFromDataURL(loaderContext, url);
      return {
        sourceURL: "",
        sourceContent
      ***REMOVED***;
    ***REMOVED***

    if (skipReading) {
      return {
        sourceURL: url
      ***REMOVED***;
    ***REMOVED***

    if (protocol === "file:") {
      const pathFromURL = _url.default.fileURLToPath(url);

      const sourceURL = _path.default.normalize(pathFromURL);

      const {
        data: sourceContent
      ***REMOVED*** = await fetchFromFilesystem(loaderContext, sourceURL);
      return {
        sourceURL,
        sourceContent
      ***REMOVED***;
    ***REMOVED***

    throw new Error(`Failed to parse source map: '${url***REMOVED***' URL is not supported`);
  ***REMOVED*** // 2. It's a scheme-relative


  if (/^\/\//.test(url)) {
    throw new Error(`Failed to parse source map: '${url***REMOVED***' URL is not supported`);
  ***REMOVED*** // 3. Absolute path


  if (_path.default.isAbsolute(url)) {
    let sourceURL = _path.default.normalize(url);

    let sourceContent;

    if (!skipReading) {
      const possibleRequests = [sourceURL];

      if (url.startsWith("/")) {
        possibleRequests.push(getAbsolutePath(context, sourceURL.slice(1), sourceRoot));
      ***REMOVED***

      const result = await fetchPathsFromFilesystem(loaderContext, possibleRequests);
      sourceURL = result.path;
      sourceContent = result.data;
    ***REMOVED***

    return {
      sourceURL,
      sourceContent
    ***REMOVED***;
  ***REMOVED*** // 4. Relative path


  const sourceURL = getAbsolutePath(context, url, sourceRoot);
  let sourceContent;

  if (!skipReading) {
    const {
      data
    ***REMOVED*** = await fetchFromFilesystem(loaderContext, sourceURL);
    sourceContent = data;
  ***REMOVED***

  return {
    sourceURL,
    sourceContent
  ***REMOVED***;
***REMOVED***
