// sisteransi/src/index.js
'use strict';

const ESC = '\x1B';
const CSI = `${ESC***REMOVED***[`;
const beep = '\u0007';

const cursor = {
  to(x, y) {
    if (!y) return `${CSI***REMOVED***${x + 1***REMOVED***G`;
    return `${CSI***REMOVED***${y + 1***REMOVED***;${x + 1***REMOVED***H`;
  ***REMOVED***,
  move(x, y) {
    let ret = '';

    if (x < 0) ret += `${CSI***REMOVED***${-x***REMOVED***D`;
    else if (x > 0) ret += `${CSI***REMOVED***${x***REMOVED***C`;

    if (y < 0) ret += `${CSI***REMOVED***${-y***REMOVED***A`;
    else if (y > 0) ret += `${CSI***REMOVED***${y***REMOVED***B`;

    return ret;
  ***REMOVED***,
  up: (count = 1) => `${CSI***REMOVED***${count***REMOVED***A`,
  down: (count = 1) => `${CSI***REMOVED***${count***REMOVED***B`,
  forward: (count = 1) => `${CSI***REMOVED***${count***REMOVED***C`,
  backward: (count = 1) => `${CSI***REMOVED***${count***REMOVED***D`,
  nextLine: (count = 1) => `${CSI***REMOVED***E`.repeat(count),
  prevLine: (count = 1) => `${CSI***REMOVED***F`.repeat(count),
  left: `${CSI***REMOVED***G`,
  hide: `${CSI***REMOVED***?25l`,
  show: `${CSI***REMOVED***?25h`,
  save: `${ESC***REMOVED***7`,
  restore: `${ESC***REMOVED***8`
***REMOVED***

const scroll = {
  up: (count = 1) => `${CSI***REMOVED***S`.repeat(count),
  down: (count = 1) => `${CSI***REMOVED***T`.repeat(count)
***REMOVED***

const erase = {
  screen: `${CSI***REMOVED***2J`,
  up: (count = 1) => `${CSI***REMOVED***1J`.repeat(count),
  down: (count = 1) => `${CSI***REMOVED***J`.repeat(count),
  line: `${CSI***REMOVED***2K`,
  lineEnd: `${CSI***REMOVED***K`,
  lineStart: `${CSI***REMOVED***1K`,
  lines(count) {
    let clear = '';
    for (let i = 0; i < count; i++)
      clear += this.line + (i < count - 1 ? cursor.up() : '');
    if (count)
      clear += cursor.left;
    return clear;
  ***REMOVED***
***REMOVED***

module.exports = { cursor, scroll, erase, beep ***REMOVED***;
