// css-tree/lib/lexer/search.js
var List = require('../common/List');

function getFirstMatchNode(matchNode) {
    if ('node' in matchNode) {
        return matchNode.node;
    ***REMOVED***

    return getFirstMatchNode(matchNode.match[0]);
***REMOVED***

function getLastMatchNode(matchNode) {
    if ('node' in matchNode) {
        return matchNode.node;
    ***REMOVED***

    return getLastMatchNode(matchNode.match[matchNode.match.length - 1]);
***REMOVED***

function matchFragments(lexer, ast, match, type, name) {
    function findFragments(matchNode) {
        if (matchNode.syntax !== null &&
            matchNode.syntax.type === type &&
            matchNode.syntax.name === name) {
            var start = getFirstMatchNode(matchNode);
            var end = getLastMatchNode(matchNode);

            lexer.syntax.walk(ast, function(node, item, list) {
                if (node === start) {
                    var nodes = new List();

                    do {
                        nodes.appendData(item.data);

                        if (item.data === end) {
                            break;
                        ***REMOVED***

                        item = item.next;
                    ***REMOVED*** while (item !== null);

                    fragments.push({
                        parent: list,
                        nodes: nodes
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        if (Array.isArray(matchNode.match)) {
            matchNode.match.forEach(findFragments);
        ***REMOVED***
    ***REMOVED***

    var fragments = [];

    if (match.matched !== null) {
        findFragments(match.matched);
    ***REMOVED***

    return fragments;
***REMOVED***

module.exports = {
    matchFragments: matchFragments
***REMOVED***;
