// @babel/plugin-transform-typescript/lib/enum.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = transpileEnum;
exports.isSyntacticallyString = isSyntacticallyString;
exports.translateEnumValues = translateEnumValues;
var _core = require("@babel/core");
var _assert = require("assert");
var _helperAnnotateAsPure = require("@babel/helper-annotate-as-pure");
var _helperSkipTransparentExpressionWrappers = require("@babel/helper-skip-transparent-expression-wrappers");
const ENUMS = new WeakMap();
const buildEnumWrapper = _core.template.expression(`
    (function (ID) {
      ASSIGNMENTS;
      return ID;
    ***REMOVED***)(INIT)
  `);
function transpileEnum(path, t) {
  const {
    node,
    parentPath
  ***REMOVED*** = path;
  if (node.declare) {
    path.remove();
    return;
  ***REMOVED***
  const name = node.id.name;
  const {
    fill,
    data,
    isPure
  ***REMOVED*** = enumFill(path, t, node.id);
  switch (parentPath.type) {
    case "BlockStatement":
    case "ExportNamedDeclaration":
    case "Program":
      {
        const isGlobal = t.isProgram(path.parent);
        const isSeen = seen(parentPath);
        let init = t.objectExpression([]);
        if (isSeen || isGlobal) {
          init = t.logicalExpression("||", t.cloneNode(fill.ID), init);
        ***REMOVED***
        const enumIIFE = buildEnumWrapper(Object.assign({***REMOVED***, fill, {
          INIT: init
        ***REMOVED***));
        if (isPure) (0, _helperAnnotateAsPure.default)(enumIIFE);
        if (isSeen) {
          const toReplace = parentPath.isExportDeclaration() ? parentPath : path;
          toReplace.replaceWith(t.expressionStatement(t.assignmentExpression("=", t.cloneNode(node.id), enumIIFE)));
        ***REMOVED*** else {
          path.scope.registerDeclaration(path.replaceWith(t.variableDeclaration(isGlobal ? "var" : "let", [t.variableDeclarator(node.id, enumIIFE)]))[0]);
        ***REMOVED***
        ENUMS.set(path.scope.getBindingIdentifier(name), data);
        break;
      ***REMOVED***
    default:
      throw new Error(`Unexpected enum parent '${path.parent.type***REMOVED***`);
  ***REMOVED***
  function seen(parentPath) {
    if (parentPath.isExportDeclaration()) {
      return seen(parentPath.parentPath);
    ***REMOVED***
    if (parentPath.getData(name)) {
      return true;
    ***REMOVED*** else {
      parentPath.setData(name, true);
      return false;
    ***REMOVED***
  ***REMOVED***
***REMOVED***
const buildStringAssignment = _core.template.statement(`
  ENUM["NAME"] = VALUE;
`);
const buildNumericAssignment = _core.template.statement(`
  ENUM[ENUM["NAME"] = VALUE] = "NAME";
`);
const buildEnumMember = (isString, options) => (isString ? buildStringAssignment : buildNumericAssignment)(options);
function enumFill(path, t, id) {
  const {
    enumValues,
    data,
    isPure
  ***REMOVED*** = translateEnumValues(path, t);
  const enumMembers = path.get("members");
  const assignments = [];
  for (let i = 0; i < enumMembers.length; i++) {
    const [memberName, memberValue] = enumValues[i];
    assignments.push(t.inheritsComments(buildEnumMember(isSyntacticallyString(memberValue), {
      ENUM: t.cloneNode(id),
      NAME: memberName,
      VALUE: memberValue
    ***REMOVED***), enumMembers[i].node));
  ***REMOVED***
  return {
    fill: {
      ID: t.cloneNode(id),
      ASSIGNMENTS: assignments
    ***REMOVED***,
    data,
    isPure
  ***REMOVED***;
***REMOVED***
function isSyntacticallyString(expr) {
  expr = (0, _helperSkipTransparentExpressionWrappers.skipTransparentExprWrapperNodes)(expr);
  switch (expr.type) {
    case "BinaryExpression":
      {
        const left = expr.left;
        const right = expr.right;
        return expr.operator === "+" && (isSyntacticallyString(left) || isSyntacticallyString(right));
      ***REMOVED***
    case "TemplateLiteral":
    case "StringLiteral":
      return true;
  ***REMOVED***
  return false;
***REMOVED***
function ReferencedIdentifier(expr, state) {
  const {
    seen,
    path,
    t
  ***REMOVED*** = state;
  const name = expr.node.name;
  if (seen.has(name)) {
    {
      for (let curScope = expr.scope; curScope !== path.scope; curScope = curScope.parent) {
        if (curScope.hasOwnBinding(name)) {
          return;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
    expr.replaceWith(t.memberExpression(t.cloneNode(path.node.id), t.cloneNode(expr.node)));
    expr.skip();
  ***REMOVED***
***REMOVED***
const enumSelfReferenceVisitor = {
  ReferencedIdentifier
***REMOVED***;
function translateEnumValues(path, t) {
  var _ENUMS$get;
  const bindingIdentifier = path.scope.getBindingIdentifier(path.node.id.name);
  const seen = (_ENUMS$get = ENUMS.get(bindingIdentifier)) != null ? _ENUMS$get : new Map();
  let constValue = -1;
  let lastName;
  let isPure = true;
  const enumMembers = path.get("members");
  const enumValues = enumMembers.map(memberPath => {
    const member = memberPath.node;
    const name = t.isIdentifier(member.id) ? member.id.name : member.id.value;
    const initializerPath = memberPath.get("initializer");
    const initializer = member.initializer;
    let value;
    if (initializer) {
      constValue = computeConstantValue(initializerPath, seen);
      if (constValue !== undefined) {
        seen.set(name, constValue);
        _assert(typeof constValue === "number" || typeof constValue === "string");
        if (constValue === Infinity || Number.isNaN(constValue)) {
          value = t.identifier(String(constValue));
        ***REMOVED*** else if (constValue === -Infinity) {
          value = t.unaryExpression("-", t.identifier("Infinity"));
        ***REMOVED*** else {
          value = t.valueToNode(constValue);
        ***REMOVED***
      ***REMOVED*** else {
        isPure && (isPure = initializerPath.isPure());
        if (initializerPath.isReferencedIdentifier()) {
          ReferencedIdentifier(initializerPath, {
            t,
            seen,
            path
          ***REMOVED***);
        ***REMOVED*** else {
          initializerPath.traverse(enumSelfReferenceVisitor, {
            t,
            seen,
            path
          ***REMOVED***);
        ***REMOVED***
        value = initializerPath.node;
        seen.set(name, undefined);
      ***REMOVED***
    ***REMOVED*** else if (typeof constValue === "number") {
      constValue += 1;
      value = t.numericLiteral(constValue);
      seen.set(name, constValue);
    ***REMOVED*** else if (typeof constValue === "string") {
      throw path.buildCodeFrameError("Enum member must have initializer.");
    ***REMOVED*** else {
      const lastRef = t.memberExpression(t.cloneNode(path.node.id), t.stringLiteral(lastName), true);
      value = t.binaryExpression("+", t.numericLiteral(1), lastRef);
      seen.set(name, undefined);
    ***REMOVED***
    lastName = name;
    return [name, value];
  ***REMOVED***);
  return {
    isPure,
    data: seen,
    enumValues
  ***REMOVED***;
***REMOVED***
function computeConstantValue(path, prevMembers, seen = new Set()) {
  return evaluate(path);
  function evaluate(path) {
    const expr = path.node;
    switch (expr.type) {
      case "MemberExpression":
        return evaluateRef(path, prevMembers, seen);
      case "StringLiteral":
        return expr.value;
      case "UnaryExpression":
        return evalUnaryExpression(path);
      case "BinaryExpression":
        return evalBinaryExpression(path);
      case "NumericLiteral":
        return expr.value;
      case "ParenthesizedExpression":
        return evaluate(path.get("expression"));
      case "Identifier":
        return evaluateRef(path, prevMembers, seen);
      case "TemplateLiteral":
        {
          if (expr.quasis.length === 1) {
            return expr.quasis[0].value.cooked;
          ***REMOVED***
          const paths = path.get("expressions");
          const quasis = expr.quasis;
          let str = "";
          for (let i = 0; i < quasis.length; i++) {
            str += quasis[i].value.cooked;
            if (i + 1 < quasis.length) {
              const value = evaluateRef(paths[i], prevMembers, seen);
              if (value === undefined) return undefined;
              str += value;
            ***REMOVED***
          ***REMOVED***
          return str;
        ***REMOVED***
      default:
        return undefined;
    ***REMOVED***
  ***REMOVED***
  function evaluateRef(path, prevMembers, seen) {
    if (path.isMemberExpression()) {
      const expr = path.node;
      const obj = expr.object;
      const prop = expr.property;
      if (!_core.types.isIdentifier(obj) || (expr.computed ? !_core.types.isStringLiteral(prop) : !_core.types.isIdentifier(prop))) {
        return;
      ***REMOVED***
      const bindingIdentifier = path.scope.getBindingIdentifier(obj.name);
      const data = ENUMS.get(bindingIdentifier);
      if (!data) return;
      return data.get(prop.computed ? prop.value : prop.name);
    ***REMOVED*** else if (path.isIdentifier()) {
      const name = path.node.name;
      if (["Infinity", "NaN"].includes(name)) {
        return Number(name);
      ***REMOVED***
      let value = prevMembers == null ? void 0 : prevMembers.get(name);
      if (value !== undefined) {
        return value;
      ***REMOVED***
      if (prevMembers != null && prevMembers.has(name)) {
        return undefined;
      ***REMOVED***
      if (seen.has(path.node)) return;
      seen.add(path.node);
      value = computeConstantValue(path.resolve(), prevMembers, seen);
      return value;
    ***REMOVED***
  ***REMOVED***
  function evalUnaryExpression(path) {
    const value = evaluate(path.get("argument"));
    if (value === undefined) {
      return undefined;
    ***REMOVED***
    switch (path.node.operator) {
      case "+":
        return value;
      case "-":
        return -value;
      case "~":
        return ~value;
      default:
        return undefined;
    ***REMOVED***
  ***REMOVED***
  function evalBinaryExpression(path) {
    const left = evaluate(path.get("left"));
    if (left === undefined) {
      return undefined;
    ***REMOVED***
    const right = evaluate(path.get("right"));
    if (right === undefined) {
      return undefined;
    ***REMOVED***
    switch (path.node.operator) {
      case "|":
        return left | right;
      case "&":
        return left & right;
      case ">>":
        return left >> right;
      case ">>>":
        return left >>> right;
      case "<<":
        return left << right;
      case "^":
        return left ^ right;
      case "*":
        return left * right;
      case "/":
        return left / right;
      case "+":
        return left + right;
      case "-":
        return left - right;
      case "%":
        return left % right;
      case "**":
        return Math.pow(left, right);
      default:
        return undefined;
    ***REMOVED***
  ***REMOVED***
***REMOVED***
