// @svgr/plugin-svgo/lib/config.js
"use strict";

exports.__esModule = true;
exports.getFilePath = getFilePath;
exports.getBaseSvgoConfig = getBaseSvgoConfig;
exports.getPlugins = getPlugins;
exports.mergeSvgoConfig = mergeSvgoConfig;

var _deepmerge = _interopRequireDefault(require("deepmerge"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj ***REMOVED***; ***REMOVED***

function getFilePath(state) {
  return state.filePath || process.cwd();
***REMOVED***

function getBaseSvgoConfig(config) {
  const baseSvgoConfig = {
    plugins: [{
      prefixIds: true
    ***REMOVED***]
  ***REMOVED***;

  if (config.icon || config.dimensions === false) {
    baseSvgoConfig.plugins.push({
      removeViewBox: false
    ***REMOVED***);
  ***REMOVED***

  return baseSvgoConfig;
***REMOVED***

function getPlugins(config) {
  if (!config || !config.plugins) {
    return [];
  ***REMOVED***

  if (!Array.isArray(config.plugins)) {
    throw Error('`svgoConfig.plugins` must be an array');
  ***REMOVED***

  return config.plugins;
***REMOVED***

function extractPlugins(config) {
  if (!config) return [];
  if (!config.plugins) return [];
  if (!Array.isArray(config.plugins)) return [config.plugins];
  return config.plugins;
***REMOVED***

function mergePlugins(configs) {
  const plugins = configs.reduce((merged, config) => _deepmerge.default.all([merged, ...extractPlugins(config)]), {***REMOVED***);
  return Object.keys(plugins).reduce((array, key) => {
    array.push({
      [key]: plugins[key]
    ***REMOVED***);
    return array;
  ***REMOVED***, []);
***REMOVED***

function mergeSvgoConfig(...configs) {
  const plugins = mergePlugins(configs);
  return { ..._deepmerge.default.all(configs.filter(Boolean)),
    plugins
  ***REMOVED***;
***REMOVED***
