// @babel/template/lib/parse.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = parseAndBuildMetadata;
var _t = require("@babel/types");
var _parser = require("@babel/parser");
var _codeFrame = require("@babel/code-frame");
const {
  isCallExpression,
  isExpressionStatement,
  isFunction,
  isIdentifier,
  isJSXIdentifier,
  isNewExpression,
  isPlaceholder,
  isStatement,
  isStringLiteral,
  removePropertiesDeep,
  traverse
***REMOVED*** = _t;
const PATTERN = /^[_$A-Z0-9]+$/;
function parseAndBuildMetadata(formatter, code, opts) {
  const {
    placeholderWhitelist,
    placeholderPattern,
    preserveComments,
    syntacticPlaceholders
  ***REMOVED*** = opts;
  const ast = parseWithCodeFrame(code, opts.parser, syntacticPlaceholders);
  removePropertiesDeep(ast, {
    preserveComments
  ***REMOVED***);
  formatter.validate(ast);
  const state = {
    syntactic: {
      placeholders: [],
      placeholderNames: new Set()
    ***REMOVED***,
    legacy: {
      placeholders: [],
      placeholderNames: new Set()
    ***REMOVED***,
    placeholderWhitelist,
    placeholderPattern,
    syntacticPlaceholders
  ***REMOVED***;
  traverse(ast, placeholderVisitorHandler, state);
  return Object.assign({
    ast
  ***REMOVED***, state.syntactic.placeholders.length ? state.syntactic : state.legacy);
***REMOVED***
function placeholderVisitorHandler(node, ancestors, state) {
  var _state$placeholderWhi;
  let name;
  let hasSyntacticPlaceholders = state.syntactic.placeholders.length > 0;
  if (isPlaceholder(node)) {
    if (state.syntacticPlaceholders === false) {
      throw new Error("%%foo%%-style placeholders can't be used when " + "'.syntacticPlaceholders' is false.");
    ***REMOVED***
    name = node.name.name;
    hasSyntacticPlaceholders = true;
  ***REMOVED*** else if (hasSyntacticPlaceholders || state.syntacticPlaceholders) {
    return;
  ***REMOVED*** else if (isIdentifier(node) || isJSXIdentifier(node)) {
    name = node.name;
  ***REMOVED*** else if (isStringLiteral(node)) {
    name = node.value;
  ***REMOVED*** else {
    return;
  ***REMOVED***
  if (hasSyntacticPlaceholders && (state.placeholderPattern != null || state.placeholderWhitelist != null)) {
    throw new Error("'.placeholderWhitelist' and '.placeholderPattern' aren't compatible" + " with '.syntacticPlaceholders: true'");
  ***REMOVED***
  if (!hasSyntacticPlaceholders && (state.placeholderPattern === false || !(state.placeholderPattern || PATTERN).test(name)) && !((_state$placeholderWhi = state.placeholderWhitelist) != null && _state$placeholderWhi.has(name))) {
    return;
  ***REMOVED***
  ancestors = ancestors.slice();
  const {
    node: parent,
    key
  ***REMOVED*** = ancestors[ancestors.length - 1];
  let type;
  if (isStringLiteral(node) || isPlaceholder(node, {
    expectedNode: "StringLiteral"
  ***REMOVED***)) {
    type = "string";
  ***REMOVED*** else if (isNewExpression(parent) && key === "arguments" || isCallExpression(parent) && key === "arguments" || isFunction(parent) && key === "params") {
    type = "param";
  ***REMOVED*** else if (isExpressionStatement(parent) && !isPlaceholder(node)) {
    type = "statement";
    ancestors = ancestors.slice(0, -1);
  ***REMOVED*** else if (isStatement(node) && isPlaceholder(node)) {
    type = "statement";
  ***REMOVED*** else {
    type = "other";
  ***REMOVED***
  const {
    placeholders,
    placeholderNames
  ***REMOVED*** = !hasSyntacticPlaceholders ? state.legacy : state.syntactic;
  placeholders.push({
    name,
    type,
    resolve: ast => resolveAncestors(ast, ancestors),
    isDuplicate: placeholderNames.has(name)
  ***REMOVED***);
  placeholderNames.add(name);
***REMOVED***
function resolveAncestors(ast, ancestors) {
  let parent = ast;
  for (let i = 0; i < ancestors.length - 1; i++) {
    const {
      key,
      index
    ***REMOVED*** = ancestors[i];
    if (index === undefined) {
      parent = parent[key];
    ***REMOVED*** else {
      parent = parent[key][index];
    ***REMOVED***
  ***REMOVED***
  const {
    key,
    index
  ***REMOVED*** = ancestors[ancestors.length - 1];
  return {
    parent,
    key,
    index
  ***REMOVED***;
***REMOVED***
function parseWithCodeFrame(code, parserOpts, syntacticPlaceholders) {
  const plugins = (parserOpts.plugins || []).slice();
  if (syntacticPlaceholders !== false) {
    plugins.push("placeholders");
  ***REMOVED***
  parserOpts = Object.assign({
    allowAwaitOutsideFunction: true,
    allowReturnOutsideFunction: true,
    allowNewTargetOutsideFunction: true,
    allowSuperOutsideMethod: true,
    allowYieldOutsideFunction: true,
    sourceType: "module"
  ***REMOVED***, parserOpts, {
    plugins
  ***REMOVED***);
  try {
    return (0, _parser.parse)(code, parserOpts);
  ***REMOVED*** catch (err) {
    const loc = err.loc;
    if (loc) {
      err.message += "\n" + (0, _codeFrame.codeFrameColumns)(code, {
        start: loc
      ***REMOVED***);
      err.code = "BABEL_TEMPLATE_PARSE_ERROR";
    ***REMOVED***
    throw err;
  ***REMOVED***
***REMOVED***
