// glob/dist/esm/index.js
import { escape, unescape ***REMOVED*** from 'minimatch';
import { Glob ***REMOVED*** from './glob.js';
import { hasMagic ***REMOVED*** from './has-magic.js';
export { escape, unescape ***REMOVED*** from 'minimatch';
export { Glob ***REMOVED*** from './glob.js';
export { hasMagic ***REMOVED*** from './has-magic.js';
export { Ignore ***REMOVED*** from './ignore.js';
export function globStreamSync(pattern, options = {***REMOVED***) {
    return new Glob(pattern, options).streamSync();
***REMOVED***
export function globStream(pattern, options = {***REMOVED***) {
    return new Glob(pattern, options).stream();
***REMOVED***
export function globSync(pattern, options = {***REMOVED***) {
    return new Glob(pattern, options).walkSync();
***REMOVED***
async function glob_(pattern, options = {***REMOVED***) {
    return new Glob(pattern, options).walk();
***REMOVED***
export function globIterateSync(pattern, options = {***REMOVED***) {
    return new Glob(pattern, options).iterateSync();
***REMOVED***
export function globIterate(pattern, options = {***REMOVED***) {
    return new Glob(pattern, options).iterate();
***REMOVED***
// aliases: glob.sync.stream() glob.stream.sync() glob.sync() etc
export const streamSync = globStreamSync;
export const stream = Object.assign(globStream, { sync: globStreamSync ***REMOVED***);
export const iterateSync = globIterateSync;
export const iterate = Object.assign(globIterate, {
    sync: globIterateSync,
***REMOVED***);
export const sync = Object.assign(globSync, {
    stream: globStreamSync,
    iterate: globIterateSync,
***REMOVED***);
export const glob = Object.assign(glob_, {
    glob: glob_,
    globSync,
    sync,
    globStream,
    stream,
    globStreamSync,
    streamSync,
    globIterate,
    iterate,
    globIterateSync,
    iterateSync,
    Glob,
    hasMagic,
    escape,
    unescape,
***REMOVED***);
glob.glob = glob;
