// eslint-plugin-react/lib/util/ast.js
/**
 * @fileoverview Utility functions for AST
 */

'use strict';

const estraverse = require('estraverse');
const eslintUtil = require('./eslint');

const getFirstTokens = eslintUtil.getFirstTokens;
const getScope = eslintUtil.getScope;
const getSourceCode = eslintUtil.getSourceCode;
// const pragmaUtil = require('./pragma');

/**
 * Wrapper for estraverse.traverse
 *
 * @param {ASTNode***REMOVED*** ASTnode The AST node being checked
 * @param {Object***REMOVED*** visitor Visitor Object for estraverse
 */
function traverse(ASTnode, visitor) {
  const opts = Object.assign({***REMOVED***, {
    fallback(node) {
      return Object.keys(node).filter((key) => key === 'children' || key === 'argument');
    ***REMOVED***,
  ***REMOVED***, visitor);

  opts.keys = Object.assign({***REMOVED***, visitor.keys, {
    JSXElement: ['children'],
    JSXFragment: ['children'],
  ***REMOVED***);

  estraverse.traverse(ASTnode, opts);
***REMOVED***

function loopNodes(nodes) {
  for (let i = nodes.length - 1; i >= 0; i--) {
    if (nodes[i].type === 'ReturnStatement') {
      return nodes[i];
    ***REMOVED***
    if (nodes[i].type === 'SwitchStatement') {
      const j = nodes[i].cases.length - 1;
      if (j >= 0) {
        return loopNodes(nodes[i].cases[j].consequent);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  return false;
***REMOVED***

/**
 * Find a return statement in the current node
 *
 * @param {ASTNode***REMOVED*** node The AST node being checked
 * @returns {ASTNode | false***REMOVED***
 */
function findReturnStatement(node) {
  if (
    (!node.value || !node.value.body || !node.value.body.body)
    && (!node.body || !node.body.body)
  ) {
    return false;
  ***REMOVED***

  const bodyNodes = node.value ? node.value.body.body : node.body.body;

  return loopNodes(bodyNodes);
***REMOVED***

// eslint-disable-next-line valid-jsdoc -- valid-jsdoc cannot parse function types.
/**
 * Helper function for traversing "returns" (return statements or the
 * returned expression in the case of an arrow function) of a function
 *
 * @param {ASTNode***REMOVED*** ASTNode The AST node being checked
 * @param {Context***REMOVED*** context The context of `ASTNode`.
 * @param {(returnValue: ASTNode, breakTraverse: () => void) => void***REMOVED*** onReturn
 *   Function to execute for each returnStatement found
 * @returns {undefined***REMOVED***
 */
function traverseReturns(ASTNode, context, onReturn) {
  const nodeType = ASTNode.type;

  if (nodeType === 'ReturnStatement') {
    onReturn(ASTNode.argument, () => {***REMOVED***);
    return;
  ***REMOVED***

  if (nodeType === 'ArrowFunctionExpression' && ASTNode.expression) {
    onReturn(ASTNode.body, () => {***REMOVED***);
    return;
  ***REMOVED***

  /* TODO: properly warn on React.forwardRefs having typo properties
  if (astUtil.isCallExpression(ASTNode)) {
    const callee = ASTNode.callee;
    const pragma = pragmaUtil.getFromContext(context);
    if (
      callee.type === 'MemberExpression'
      && callee.object.type === 'Identifier'
      && callee.object.name === pragma
      && callee.property.type === 'Identifier'
      && callee.property.name === 'forwardRef'
      && ASTNode.arguments.length > 0
    ) {
      return enterFunc(ASTNode.arguments[0]);
    ***REMOVED***
    return;
  ***REMOVED***
  */

  if (
    nodeType !== 'FunctionExpression'
    && nodeType !== 'FunctionDeclaration'
    && nodeType !== 'ArrowFunctionExpression'
    && nodeType !== 'MethodDefinition'
  ) {
    return;
  ***REMOVED***

  traverse(ASTNode.body, {
    enter(node) {
      const breakTraverse = () => {
        this.break();
      ***REMOVED***;
      switch (node.type) {
        case 'ReturnStatement':
          this.skip();
          onReturn(node.argument, breakTraverse);
          return;
        case 'BlockStatement':
        case 'IfStatement':
        case 'ForStatement':
        case 'WhileStatement':
        case 'SwitchStatement':
        case 'SwitchCase':
          return;
        default:
          this.skip();
      ***REMOVED***
    ***REMOVED***,
  ***REMOVED***);
***REMOVED***

/**
 * Get node with property's name
 * @param {Object***REMOVED*** node - Property.
 * @returns {Object***REMOVED*** Property name node.
 */
function getPropertyNameNode(node) {
  if (
    node.key
    || node.type === 'MethodDefinition'
    || node.type === 'Property'
  ) {
    return node.key;
  ***REMOVED***
  if (node.type === 'MemberExpression') {
    return node.property;
  ***REMOVED***
  return null;
***REMOVED***

/**
 * Get properties name
 * @param {Object***REMOVED*** node - Property.
 * @returns {string***REMOVED*** Property name.
 */
function getPropertyName(node) {
  const nameNode = getPropertyNameNode(node);
  return nameNode ? nameNode.name : '';
***REMOVED***

/**
 * Get properties for a given AST node
 * @param {ASTNode***REMOVED*** node The AST node being checked.
 * @returns {Array***REMOVED*** Properties array.
 */
function getComponentProperties(node) {
  switch (node.type) {
    case 'ClassDeclaration':
    case 'ClassExpression':
      return node.body.body;
    case 'ObjectExpression':
      return node.properties;
    default:
      return [];
  ***REMOVED***
***REMOVED***

/**
 * Gets the first node in a line from the initial node, excluding whitespace.
 * @param {Object***REMOVED*** context The node to check
 * @param {ASTNode***REMOVED*** node The node to check
 * @return {ASTNode***REMOVED*** the first node in the line
 */
function getFirstNodeInLine(context, node) {
  const sourceCode = getSourceCode(context);
  let token = node;
  let lines;
  do {
    token = sourceCode.getTokenBefore(token);
    lines = token.type === 'JSXText'
      ? token.value.split('\n')
      : null;
  ***REMOVED*** while (
    token.type === 'JSXText'
        && /^\s*$/.test(lines[lines.length - 1])
  );
  return token;
***REMOVED***

/**
 * Checks if the node is the first in its line, excluding whitespace.
 * @param {Object***REMOVED*** context The node to check
 * @param {ASTNode***REMOVED*** node The node to check
 * @return {boolean***REMOVED*** true if it's the first node in its line
 */
function isNodeFirstInLine(context, node) {
  const token = getFirstNodeInLine(context, node);
  const startLine = node.loc.start.line;
  const endLine = token ? token.loc.end.line : -1;
  return startLine !== endLine;
***REMOVED***

/**
 * Checks if the node is a function or arrow function expression.
 * @param {ASTNode***REMOVED*** node The node to check
 * @return {boolean***REMOVED*** true if it's a function-like expression
 */
function isFunctionLikeExpression(node) {
  return node.type === 'FunctionExpression' || node.type === 'ArrowFunctionExpression';
***REMOVED***

/**
 * Checks if the node is a function.
 * @param {ASTNode***REMOVED*** node The node to check
 * @return {boolean***REMOVED*** true if it's a function
 */
function isFunction(node) {
  return node.type === 'FunctionExpression' || node.type === 'FunctionDeclaration';
***REMOVED***

/**
 * Checks if node is a function declaration or expression or arrow function.
 * @param {ASTNode***REMOVED*** node The node to check
 * @return {boolean***REMOVED*** true if it's a function-like
 */
function isFunctionLike(node) {
  return node.type === 'FunctionDeclaration' || isFunctionLikeExpression(node);
***REMOVED***

/**
 * Checks if the node is a class.
 * @param {ASTNode***REMOVED*** node The node to check
 * @return {boolean***REMOVED*** true if it's a class
 */
function isClass(node) {
  return node.type === 'ClassDeclaration' || node.type === 'ClassExpression';
***REMOVED***

/**
 * Check if we are in a class constructor
 * @param {Context***REMOVED*** context
 * @param {ASTNode***REMOVED*** node The AST node being checked.
 * @return {boolean***REMOVED***
 */
function inConstructor(context, node) {
  let scope = getScope(context, node);
  while (scope) {
    // @ts-ignore
    if (scope.block && scope.block.parent && scope.block.parent.kind === 'constructor') {
      return true;
    ***REMOVED***
    scope = scope.upper;
  ***REMOVED***
  return false;
***REMOVED***

/**
 * Removes quotes from around an identifier.
 * @param {string***REMOVED*** string the identifier to strip
 * @returns {string***REMOVED***
 */
function stripQuotes(string) {
  return string.replace(/^'|'$/g, '');
***REMOVED***

/**
 * Retrieve the name of a key node
 * @param {Context***REMOVED*** context The AST node with the key.
 * @param {any***REMOVED*** node The AST node with the key.
 * @return {string | undefined***REMOVED*** the name of the key
 */
function getKeyValue(context, node) {
  if (node.type === 'ObjectTypeProperty') {
    const tokens = getFirstTokens(context, node, 2);
    return (tokens[0].value === '+' || tokens[0].value === '-'
      ? tokens[1].value
      : stripQuotes(tokens[0].value)
    );
  ***REMOVED***
  if (node.type === 'GenericTypeAnnotation') {
    return node.id.name;
  ***REMOVED***
  if (node.type === 'ObjectTypeAnnotation') {
    return;
  ***REMOVED***
  const key = node.key || node.argument;
  if (!key) {
    return;
  ***REMOVED***
  return key.type === 'Identifier' ? key.name : key.value;
***REMOVED***

/**
 * Checks if a node is surrounded by parenthesis.
 *
 * @param {object***REMOVED*** context - Context from the rule
 * @param {ASTNode***REMOVED*** node - Node to be checked
 * @returns {boolean***REMOVED***
 */
function isParenthesized(context, node) {
  const sourceCode = getSourceCode(context);
  const previousToken = sourceCode.getTokenBefore(node);
  const nextToken = sourceCode.getTokenAfter(node);

  return !!previousToken && !!nextToken
    && previousToken.value === '(' && previousToken.range[1] <= node.range[0]
    && nextToken.value === ')' && nextToken.range[0] >= node.range[1];
***REMOVED***

/**
 * Checks if a node is being assigned a value: props.bar = 'bar'
 * @param {ASTNode***REMOVED*** node The AST node being checked.
 * @returns {boolean***REMOVED***
 */
function isAssignmentLHS(node) {
  return (
    node.parent
    && node.parent.type === 'AssignmentExpression'
    && node.parent.left === node
  );
***REMOVED***

function isTSAsExpression(node) {
  return node && node.type === 'TSAsExpression';
***REMOVED***

/**
 * Matcher used to check whether given node is a `CallExpression`
 * @param {ASTNode***REMOVED*** node The AST node
 * @returns {boolean***REMOVED*** True if node is a `CallExpression`, false if not
 */
function isCallExpression(node) {
  return node && node.type === 'CallExpression';
***REMOVED***

/**
 * Extracts the expression node that is wrapped inside a TS type assertion
 *
 * @param {ASTNode***REMOVED*** node - potential TS node
 * @returns {ASTNode***REMOVED*** - unwrapped expression node
 */
function unwrapTSAsExpression(node) {
  return isTSAsExpression(node) ? node.expression : node;
***REMOVED***

function isTSTypeReference(node) {
  if (!node) return false;

  return node.type === 'TSTypeReference';
***REMOVED***

function isTSTypeAnnotation(node) {
  if (!node) { return false; ***REMOVED***

  return node.type === 'TSTypeAnnotation';
***REMOVED***

function isTSTypeLiteral(node) {
  if (!node) { return false; ***REMOVED***

  return node.type === 'TSTypeLiteral';
***REMOVED***

function isTSIntersectionType(node) {
  if (!node) { return false; ***REMOVED***

  return node.type === 'TSIntersectionType';
***REMOVED***

function isTSInterfaceHeritage(node) {
  if (!node) { return false; ***REMOVED***

  return node.type === 'TSInterfaceHeritage';
***REMOVED***

function isTSInterfaceDeclaration(node) {
  if (!node) { return false; ***REMOVED***

  return (node.type === 'ExportNamedDeclaration' && node.declaration
    ? node.declaration.type
    : node.type
  ) === 'TSInterfaceDeclaration';
***REMOVED***

function isTSTypeDeclaration(node) {
  if (!node) { return false; ***REMOVED***

  const nodeToCheck = node.type === 'ExportNamedDeclaration' && node.declaration
    ? node.declaration
    : node;

  return nodeToCheck.type === 'VariableDeclaration' && nodeToCheck.kind === 'type';
***REMOVED***

function isTSTypeAliasDeclaration(node) {
  if (!node) { return false; ***REMOVED***

  if (node.type === 'ExportNamedDeclaration' && node.declaration) {
    return node.declaration.type === 'TSTypeAliasDeclaration' && node.exportKind === 'type';
  ***REMOVED***
  return node.type === 'TSTypeAliasDeclaration';
***REMOVED***

function isTSParenthesizedType(node) {
  if (!node) { return false; ***REMOVED***

  return node.type === 'TSTypeAliasDeclaration';
***REMOVED***

function isTSFunctionType(node) {
  if (!node) { return false; ***REMOVED***

  return node.type === 'TSFunctionType';
***REMOVED***

function isTSTypeQuery(node) {
  if (!node) { return false; ***REMOVED***

  return node.type === 'TSTypeQuery';
***REMOVED***

function isTSTypeParameterInstantiation(node) {
  if (!node) { return false; ***REMOVED***

  return node.type === 'TSTypeParameterInstantiation';
***REMOVED***

module.exports = {
  findReturnStatement,
  getComponentProperties,
  getFirstNodeInLine,
  getKeyValue,
  getPropertyName,
  getPropertyNameNode,
  inConstructor,
  isAssignmentLHS,
  isCallExpression,
  isClass,
  isFunction,
  isFunctionLike,
  isFunctionLikeExpression,
  isNodeFirstInLine,
  isParenthesized,
  isTSAsExpression,
  isTSFunctionType,
  isTSInterfaceDeclaration,
  isTSInterfaceHeritage,
  isTSIntersectionType,
  isTSParenthesizedType,
  isTSTypeAliasDeclaration,
  isTSTypeAnnotation,
  isTSTypeDeclaration,
  isTSTypeLiteral,
  isTSTypeParameterInstantiation,
  isTSTypeQuery,
  isTSTypeReference,
  traverse,
  traverseReturns,
  unwrapTSAsExpression,
***REMOVED***;
