// eslint-plugin-react/lib/util/Components.js
/**
 * @fileoverview Utility class and functions for React components detection
 * @author Yannick Croissant
 */

'use strict';

const arrayIncludes = require('array-includes');
const fromEntries = require('object.fromentries');
const values = require('object.values');
const iterFrom = require('es-iterator-helpers/Iterator.from');
const map = require('es-iterator-helpers/Iterator.prototype.map');

const variableUtil = require('./variable');
const pragmaUtil = require('./pragma');
const astUtil = require('./ast');
const componentUtil = require('./componentUtil');
const propTypesUtil = require('./propTypes');
const jsxUtil = require('./jsx');
const usedPropTypesUtil = require('./usedPropTypes');
const defaultPropsUtil = require('./defaultProps');
const isFirstLetterCapitalized = require('./isFirstLetterCapitalized');
const isDestructuredFromPragmaImport = require('./isDestructuredFromPragmaImport');
const eslintUtil = require('./eslint');

const getScope = eslintUtil.getScope;
const getText = eslintUtil.getText;

function getId(node) {
  return node ? `${node.range[0]***REMOVED***:${node.range[1]***REMOVED***` : '';
***REMOVED***

function usedPropTypesAreEquivalent(propA, propB) {
  if (propA.name === propB.name) {
    if (!propA.allNames && !propB.allNames) {
      return true;
    ***REMOVED***
    if (Array.isArray(propA.allNames) && Array.isArray(propB.allNames) && propA.allNames.join('') === propB.allNames.join('')) {
      return true;
    ***REMOVED***
    return false;
  ***REMOVED***
  return false;
***REMOVED***

function mergeUsedPropTypes(propsList, newPropsList) {
  const propsToAdd = newPropsList.filter((newProp) => {
    const newPropIsAlreadyInTheList = propsList.some((prop) => usedPropTypesAreEquivalent(prop, newProp));
    return !newPropIsAlreadyInTheList;
  ***REMOVED***);

  return propsList.concat(propsToAdd);
***REMOVED***

const USE_HOOK_PREFIX_REGEX = /^use[A-Z]/;

const Lists = new WeakMap();
const ReactImports = new WeakMap();

/**
 * Components
 */
class Components {
  constructor() {
    Lists.set(this, {***REMOVED***);
    ReactImports.set(this, {***REMOVED***);
  ***REMOVED***

  /**
   * Add a node to the components list, or update it if it's already in the list
   *
   * @param {ASTNode***REMOVED*** node The AST node being added.
   * @param {number***REMOVED*** confidence Confidence in the component detection (0=banned, 1=maybe, 2=yes)
   * @returns {Object***REMOVED*** Added component object
   */
  add(node, confidence) {
    const id = getId(node);
    const list = Lists.get(this);
    if (list[id]) {
      if (confidence === 0 || list[id].confidence === 0) {
        list[id].confidence = 0;
      ***REMOVED*** else {
        list[id].confidence = Math.max(list[id].confidence, confidence);
      ***REMOVED***
      return list[id];
    ***REMOVED***
    list[id] = {
      node,
      confidence,
    ***REMOVED***;
    return list[id];
  ***REMOVED***

  /**
   * Find a component in the list using its node
   *
   * @param {ASTNode***REMOVED*** node The AST node being searched.
   * @returns {Object***REMOVED*** Component object, undefined if the component is not found or has confidence value of 0.
   */
  get(node) {
    const id = getId(node);
    const item = Lists.get(this)[id];
    if (item && item.confidence >= 1) {
      return item;
    ***REMOVED***
    return null;
  ***REMOVED***

  /**
   * Update a component in the list
   *
   * @param {ASTNode***REMOVED*** node The AST node being updated.
   * @param {Object***REMOVED*** props Additional properties to add to the component.
   */
  set(node, props) {
    const list = Lists.get(this);
    let component = list[getId(node)];
    while (!component || component.confidence < 1) {
      node = node.parent;
      if (!node) {
        return;
      ***REMOVED***
      component = list[getId(node)];
    ***REMOVED***

    Object.assign(
      component,
      props,
      {
        usedPropTypes: mergeUsedPropTypes(
          component.usedPropTypes || [],
          props.usedPropTypes || []
        ),
      ***REMOVED***
    );
  ***REMOVED***

  /**
   * Return the components list
   * Components for which we are not confident are not returned
   *
   * @returns {Object***REMOVED*** Components list
   */
  list() {
    const thisList = Lists.get(this);
    const list = {***REMOVED***;
    const usedPropTypes = {***REMOVED***;

    // Find props used in components for which we are not confident
    Object.keys(thisList).filter((i) => thisList[i].confidence < 2).forEach((i) => {
      let component = null;
      let node = null;
      node = thisList[i].node;
      while (!component && node.parent) {
        node = node.parent;
        // Stop moving up if we reach a decorator
        if (node.type === 'Decorator') {
          break;
        ***REMOVED***
        component = this.get(node);
      ***REMOVED***
      if (component) {
        const newUsedProps = (thisList[i].usedPropTypes || []).filter((propType) => !propType.node || propType.node.kind !== 'init');

        const componentId = getId(component.node);

        usedPropTypes[componentId] = mergeUsedPropTypes(usedPropTypes[componentId] || [], newUsedProps);
      ***REMOVED***
    ***REMOVED***);

    // Assign used props in not confident components to the parent component
    Object.keys(thisList).filter((j) => thisList[j].confidence >= 2).forEach((j) => {
      const id = getId(thisList[j].node);
      list[j] = thisList[j];
      if (usedPropTypes[id]) {
        list[j].usedPropTypes = mergeUsedPropTypes(list[j].usedPropTypes || [], usedPropTypes[id]);
      ***REMOVED***
    ***REMOVED***);
    return list;
  ***REMOVED***

  /**
   * Return the length of the components list
   * Components for which we are not confident are not counted
   *
   * @returns {number***REMOVED*** Components list length
   */
  length() {
    const list = Lists.get(this);
    return values(list).filter((component) => component.confidence >= 2).length;
  ***REMOVED***

  /**
   * Return the node naming the default React import
   * It can be used to determine the local name of import, even if it's imported
   * with an unusual name.
   *
   * @returns {ASTNode***REMOVED*** React default import node
   */
  getDefaultReactImports() {
    return ReactImports.get(this).defaultReactImports;
  ***REMOVED***

  /**
   * Return the nodes of all React named imports
   *
   * @returns {Object***REMOVED*** The list of React named imports
   */
  getNamedReactImports() {
    return ReactImports.get(this).namedReactImports;
  ***REMOVED***

  /**
   * Add the default React import specifier to the scope
   *
   * @param {ASTNode***REMOVED*** specifier The AST Node of the default React import
   * @returns {void***REMOVED***
   */
  addDefaultReactImport(specifier) {
    const info = ReactImports.get(this);
    ReactImports.set(this, Object.assign({***REMOVED***, info, {
      defaultReactImports: (info.defaultReactImports || []).concat(specifier),
    ***REMOVED***));
  ***REMOVED***

  /**
   * Add a named React import specifier to the scope
   *
   * @param {ASTNode***REMOVED*** specifier The AST Node of a named React import
   * @returns {void***REMOVED***
   */
  addNamedReactImport(specifier) {
    const info = ReactImports.get(this);
    ReactImports.set(this, Object.assign({***REMOVED***, info, {
      namedReactImports: (info.namedReactImports || []).concat(specifier),
    ***REMOVED***));
  ***REMOVED***
***REMOVED***

function getWrapperFunctions(context, pragma) {
  const componentWrapperFunctions = context.settings.componentWrapperFunctions || [];

  // eslint-disable-next-line arrow-body-style
  return componentWrapperFunctions.map((wrapperFunction) => {
    return typeof wrapperFunction === 'string'
      ? { property: wrapperFunction ***REMOVED***
      : Object.assign({***REMOVED***, wrapperFunction, {
        object: wrapperFunction.object === '<pragma>' ? pragma : wrapperFunction.object,
      ***REMOVED***);
  ***REMOVED***).concat([
    { property: 'forwardRef', object: pragma ***REMOVED***,
    { property: 'memo', object: pragma ***REMOVED***,
  ]);
***REMOVED***

// eslint-disable-next-line valid-jsdoc
/**
 * Merge many eslint rules into one
 * @param {{[_: string]: Function***REMOVED***[]***REMOVED*** rules the returned values for eslint rule.create(context)
 * @returns {{[_: string]: Function***REMOVED******REMOVED*** merged rule
 */
function mergeRules(rules) {
  /** @type {Map<string, Function[]>***REMOVED*** */
  const handlersByKey = new Map();
  rules.forEach((rule) => {
    Object.keys(rule).forEach((key) => {
      const fns = handlersByKey.get(key);
      if (!fns) {
        handlersByKey.set(key, [rule[key]]);
      ***REMOVED*** else {
        fns.push(rule[key]);
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***);

  /** @type {{ [key: string]: Function ***REMOVED******REMOVED*** */
  return fromEntries(map(iterFrom(handlersByKey), (entry) => [
    entry[0],
    function mergedHandler(node) {
      entry[1].forEach((fn) => {
        fn(node);
      ***REMOVED***);
    ***REMOVED***,
  ]));
***REMOVED***

function componentRule(rule, context) {
  const pragma = pragmaUtil.getFromContext(context);
  const components = new Components();
  const wrapperFunctions = getWrapperFunctions(context, pragma);

  // Utilities for component detection
  const utils = {
    /**
     * Check if variable is destructured from pragma import
     *
     * @param {ASTNode***REMOVED*** node The AST node to check
     * @param {string***REMOVED*** variable The variable name to check
     * @returns {boolean***REMOVED*** True if createElement is destructured from the pragma
     */
    isDestructuredFromPragmaImport(node, variable) {
      return isDestructuredFromPragmaImport(context, node, variable);
    ***REMOVED***,

    /**
     * @param {ASTNode***REMOVED*** node
     * @param {boolean=***REMOVED*** strict
     * @returns {boolean***REMOVED***
     */
    isReturningJSX(node, strict) {
      return jsxUtil.isReturningJSX(context, node, strict, true);
    ***REMOVED***,

    isReturningJSXOrNull(node, strict) {
      return jsxUtil.isReturningJSX(context, node, strict);
    ***REMOVED***,

    isReturningOnlyNull(node) {
      return jsxUtil.isReturningOnlyNull(node, context);
    ***REMOVED***,

    getPragmaComponentWrapper(node) {
      let isPragmaComponentWrapper;
      let currentNode = node;
      let prevNode;
      do {
        currentNode = currentNode.parent;
        isPragmaComponentWrapper = this.isPragmaComponentWrapper(currentNode);
        if (isPragmaComponentWrapper) {
          prevNode = currentNode;
        ***REMOVED***
      ***REMOVED*** while (isPragmaComponentWrapper);

      return prevNode;
    ***REMOVED***,

    getComponentNameFromJSXElement(node) {
      if (node.type !== 'JSXElement') {
        return null;
      ***REMOVED***
      if (node.openingElement && node.openingElement.name && node.openingElement.name.name) {
        return node.openingElement.name.name;
      ***REMOVED***
      return null;
    ***REMOVED***,

    /**
     * Getting the first JSX element's name.
     * @param {object***REMOVED*** node
     * @returns {string | null***REMOVED***
     */
    getNameOfWrappedComponent(node) {
      if (node.length < 1) {
        return null;
      ***REMOVED***
      const body = node[0].body;
      if (!body) {
        return null;
      ***REMOVED***
      if (body.type === 'JSXElement') {
        return this.getComponentNameFromJSXElement(body);
      ***REMOVED***
      if (body.type === 'BlockStatement') {
        const jsxElement = body.body.find((item) => item.type === 'ReturnStatement');
        return jsxElement
          && jsxElement.argument
          && this.getComponentNameFromJSXElement(jsxElement.argument);
      ***REMOVED***
      return null;
    ***REMOVED***,

    /**
     * Get the list of names of components created till now
     * @returns {string | boolean***REMOVED***
     */
    getDetectedComponents() {
      const list = components.list();
      return values(list).filter((val) => {
        if (val.node.type === 'ClassDeclaration') {
          return true;
        ***REMOVED***
        if (
          val.node.type === 'ArrowFunctionExpression'
          && val.node.parent
          && val.node.parent.type === 'VariableDeclarator'
          && val.node.parent.id
        ) {
          return true;
        ***REMOVED***
        return false;
      ***REMOVED***).map((val) => {
        if (val.node.type === 'ArrowFunctionExpression') return val.node.parent.id.name;
        return val.node.id && val.node.id.name;
      ***REMOVED***);
    ***REMOVED***,

    /**
     * It will check whether memo/forwardRef is wrapping existing component or
     * creating a new one.
     * @param {object***REMOVED*** node
     * @returns {boolean***REMOVED***
     */
    nodeWrapsComponent(node) {
      const childComponent = this.getNameOfWrappedComponent(node.arguments);
      const componentList = this.getDetectedComponents();
      return !!childComponent && arrayIncludes(componentList, childComponent);
    ***REMOVED***,

    isPragmaComponentWrapper(node) {
      if (!astUtil.isCallExpression(node)) {
        return false;
      ***REMOVED***

      return wrapperFunctions.some((wrapperFunction) => {
        if (node.callee.type === 'MemberExpression') {
          return wrapperFunction.object
            && wrapperFunction.object === node.callee.object.name
            && wrapperFunction.property === node.callee.property.name
            && !this.nodeWrapsComponent(node);
        ***REMOVED***
        return wrapperFunction.property === node.callee.name
          && (!wrapperFunction.object
            // Functions coming from the current pragma need special handling
            || (wrapperFunction.object === pragma && this.isDestructuredFromPragmaImport(node, node.callee.name))
          );
      ***REMOVED***);
    ***REMOVED***,

    /**
     * Find a return statement in the current node
     *
     * @param {ASTNode***REMOVED*** node The AST node being checked
     */
    findReturnStatement: astUtil.findReturnStatement,

    /**
     * Get the parent component node from the current scope
     * @param {ASTNode***REMOVED*** node
     *
     * @returns {ASTNode***REMOVED*** component node, null if we are not in a component
     */
    getParentComponent(node) {
      return (
        componentUtil.getParentES6Component(context, node)
        || componentUtil.getParentES5Component(context, node)
        || utils.getParentStatelessComponent(node)
      );
    ***REMOVED***,

    /**
     * @param {ASTNode***REMOVED*** node
     * @returns {boolean***REMOVED***
     */
    isInAllowedPositionForComponent(node) {
      switch (node.parent.type) {
        case 'VariableDeclarator':
        case 'AssignmentExpression':
        case 'Property':
        case 'ReturnStatement':
        case 'ExportDefaultDeclaration':
        case 'ArrowFunctionExpression': {
          return true;
        ***REMOVED***
        case 'SequenceExpression': {
          return utils.isInAllowedPositionForComponent(node.parent)
            && node === node.parent.expressions[node.parent.expressions.length - 1];
        ***REMOVED***
        default:
          return false;
      ***REMOVED***
    ***REMOVED***,

    /**
     * Get node if node is a stateless component, or node.parent in cases like
     * `React.memo` or `React.forwardRef`. Otherwise returns `undefined`.
     * @param {ASTNode***REMOVED*** node
     * @returns {ASTNode | undefined***REMOVED***
     */
    getStatelessComponent(node) {
      const parent = node.parent;
      if (
        node.type === 'FunctionDeclaration'
        && (!node.id || isFirstLetterCapitalized(node.id.name))
        && utils.isReturningJSXOrNull(node)
      ) {
        return node;
      ***REMOVED***

      if (node.type === 'FunctionExpression' || node.type === 'ArrowFunctionExpression') {
        const isPropertyAssignment = parent.type === 'AssignmentExpression'
          && parent.left.type === 'MemberExpression';
        const isModuleExportsAssignment = isPropertyAssignment
          && parent.left.object.name === 'module'
          && parent.left.property.name === 'exports';

        if (node.parent.type === 'ExportDefaultDeclaration') {
          if (utils.isReturningJSX(node)) {
            return node;
          ***REMOVED***
          return undefined;
        ***REMOVED***

        if (node.parent.type === 'VariableDeclarator' && utils.isReturningJSXOrNull(node)) {
          if (isFirstLetterCapitalized(node.parent.id.name)) {
            return node;
          ***REMOVED***
          return undefined;
        ***REMOVED***

        // case: const any = () => { return (props) => null ***REMOVED***
        // case: const any = () => (props) => null
        if (
          (node.parent.type === 'ReturnStatement' || (node.parent.type === 'ArrowFunctionExpression' && node.parent.expression))
          && !utils.isReturningJSX(node)
        ) {
          return undefined;
        ***REMOVED***

        // case: any = () => { return => null ***REMOVED***
        // case: any = () => null
        if (node.parent.type === 'AssignmentExpression' && !isPropertyAssignment && utils.isReturningJSXOrNull(node)) {
          if (isFirstLetterCapitalized(node.parent.left.name)) {
            return node;
          ***REMOVED***
          return undefined;
        ***REMOVED***

        // case: any = () => () => null
        if (node.parent.type === 'ArrowFunctionExpression' && node.parent.parent.type === 'AssignmentExpression' && !isPropertyAssignment && utils.isReturningJSXOrNull(node)) {
          if (isFirstLetterCapitalized(node.parent.parent.left.name)) {
            return node;
          ***REMOVED***
          return undefined;
        ***REMOVED***

        // case: { any: () => () => null ***REMOVED***
        if (node.parent.type === 'ArrowFunctionExpression' && node.parent.parent.type === 'Property' && !isPropertyAssignment && utils.isReturningJSXOrNull(node)) {
          if (isFirstLetterCapitalized(node.parent.parent.key.name)) {
            return node;
          ***REMOVED***
          return undefined;
        ***REMOVED***

        // case: any = function() {return function() {return null;***REMOVED***;***REMOVED***
        if (node.parent.type === 'ReturnStatement') {
          if (isFirstLetterCapitalized(node.id && node.id.name)) {
            return node;
          ***REMOVED***
          const functionExpr = node.parent.parent.parent;
          if (functionExpr.parent.type === 'AssignmentExpression' && !isPropertyAssignment && utils.isReturningJSXOrNull(node)) {
            if (isFirstLetterCapitalized(functionExpr.parent.left.name)) {
              return node;
            ***REMOVED***
            return undefined;
          ***REMOVED***
        ***REMOVED***

        // case: { any: function() {return function() {return null;***REMOVED***;***REMOVED*** ***REMOVED***
        if (node.parent.type === 'ReturnStatement') {
          const functionExpr = node.parent.parent.parent;
          if (functionExpr.parent.type === 'Property' && !isPropertyAssignment && utils.isReturningJSXOrNull(node)) {
            if (isFirstLetterCapitalized(functionExpr.parent.key.name)) {
              return node;
            ***REMOVED***
            return undefined;
          ***REMOVED***
        ***REMOVED***

        // for case abc = { [someobject.somekey]: props => { ... return not-jsx ***REMOVED*** ***REMOVED***
        if (
          node.parent
          && node.parent.key
          && node.parent.key.type === 'MemberExpression'
          && !utils.isReturningJSX(node)
          && !utils.isReturningOnlyNull(node)
        ) {
          return undefined;
        ***REMOVED***

        if (
          node.parent.type === 'Property' && (
            (node.parent.method && !node.parent.computed) // case: { f() { return ... ***REMOVED*** ***REMOVED***
            || (!node.id && !node.parent.computed) // case: { f: () => ... ***REMOVED***
          )
        ) {
          if (
            isFirstLetterCapitalized(node.parent.key.name)
            && utils.isReturningJSX(node)
          ) {
            return node;
          ***REMOVED***
          return undefined;
        ***REMOVED***

        // Case like `React.memo(() => <></>)` or `React.forwardRef(...)`
        const pragmaComponentWrapper = utils.getPragmaComponentWrapper(node);
        if (pragmaComponentWrapper && utils.isReturningJSXOrNull(node)) {
          return pragmaComponentWrapper;
        ***REMOVED***

        if (!(utils.isInAllowedPositionForComponent(node) && utils.isReturningJSXOrNull(node))) {
          return undefined;
        ***REMOVED***

        if (utils.isParentComponentNotStatelessComponent(node)) {
          return undefined;
        ***REMOVED***

        if (node.id) {
          return isFirstLetterCapitalized(node.id.name) ? node : undefined;
        ***REMOVED***

        if (
          isPropertyAssignment
          && !isModuleExportsAssignment
          && !isFirstLetterCapitalized(parent.left.property.name)
        ) {
          return undefined;
        ***REMOVED***

        if (parent.type === 'Property' && utils.isReturningOnlyNull(node)) {
          return undefined;
        ***REMOVED***

        return node;
      ***REMOVED***

      return undefined;
    ***REMOVED***,

    /**
     * Get the parent stateless component node from the current scope
     *
     * @param {ASTNode***REMOVED*** node The AST node being checked
     * @returns {ASTNode***REMOVED*** component node, null if we are not in a component
     */
    getParentStatelessComponent(node) {
      let scope = getScope(context, node);
      while (scope) {
        const statelessComponent = utils.getStatelessComponent(scope.block);
        if (statelessComponent) {
          return statelessComponent;
        ***REMOVED***
        scope = scope.upper;
      ***REMOVED***
      return null;
    ***REMOVED***,

    /**
     * Get the related component from a node
     *
     * @param {ASTNode***REMOVED*** node The AST node being checked (must be a MemberExpression).
     * @returns {ASTNode | null***REMOVED*** component node, null if we cannot find the component
     */
    getRelatedComponent(node) {
      let i;
      let j;
      let k;
      let l;
      let componentNode;
      // Get the component path
      const componentPath = [];
      let nodeTemp = node;
      while (nodeTemp) {
        if (nodeTemp.property && nodeTemp.property.type === 'Identifier') {
          componentPath.push(nodeTemp.property.name);
        ***REMOVED***
        if (nodeTemp.object && nodeTemp.object.type === 'Identifier') {
          componentPath.push(nodeTemp.object.name);
        ***REMOVED***
        nodeTemp = nodeTemp.object;
      ***REMOVED***
      componentPath.reverse();
      const componentName = componentPath.slice(0, componentPath.length - 1).join('.');

      // Find the variable in the current scope
      const variableName = componentPath.shift();
      if (!variableName) {
        return null;
      ***REMOVED***
      const variableInScope = variableUtil.getVariableFromContext(context, node, variableName);
      if (!variableInScope) {
        return null;
      ***REMOVED***

      // Try to find the component using variable references
      variableInScope.references.some((ref) => {
        let refId = ref.identifier;
        if (refId.parent && refId.parent.type === 'MemberExpression') {
          refId = refId.parent;
        ***REMOVED***
        if (getText(context, refId) !== componentName) {
          return false;
        ***REMOVED***
        if (refId.type === 'MemberExpression') {
          componentNode = refId.parent.right;
        ***REMOVED*** else if (
          refId.parent
          && refId.parent.type === 'VariableDeclarator'
          && refId.parent.init
          && refId.parent.init.type !== 'Identifier'
        ) {
          componentNode = refId.parent.init;
        ***REMOVED***
        return true;
      ***REMOVED***);

      if (componentNode) {
        // Return the component
        return components.add(componentNode, 1);
      ***REMOVED***

      // Try to find the component using variable declarations
      const defs = variableInScope.defs;
      const defInScope = defs.find((def) => (
        def.type === 'ClassName'
        || def.type === 'FunctionName'
        || def.type === 'Variable'
      ));
      if (!defInScope || !defInScope.node) {
        return null;
      ***REMOVED***
      componentNode = defInScope.node.init || defInScope.node;

      // Traverse the node properties to the component declaration
      for (i = 0, j = componentPath.length; i < j; i++) {
        if (!componentNode.properties) {
          continue; // eslint-disable-line no-continue
        ***REMOVED***
        for (k = 0, l = componentNode.properties.length; k < l; k++) {
          if (componentNode.properties[k].key && componentNode.properties[k].key.name === componentPath[i]) {
            componentNode = componentNode.properties[k];
            break;
          ***REMOVED***
        ***REMOVED***
        if (!componentNode || !componentNode.value) {
          return null;
        ***REMOVED***
        componentNode = componentNode.value;
      ***REMOVED***

      // Return the component
      return components.add(componentNode, 1);
    ***REMOVED***,

    isParentComponentNotStatelessComponent(node) {
      return !!(
        node.parent
        && node.parent.key
        && node.parent.key.type === 'Identifier'
        // custom component functions must start with a capital letter (returns false otherwise)
        && node.parent.key.name.charAt(0) === node.parent.key.name.charAt(0).toLowerCase()
        // react render function cannot have params
        && !!(node.params || []).length
      );
    ***REMOVED***,

    /**
     * Identify whether a node (CallExpression) is a call to a React hook
     *
     * @param {ASTNode***REMOVED*** node The AST node being searched. (expects CallExpression)
     * @param {('useCallback'|'useContext'|'useDebugValue'|'useEffect'|'useImperativeHandle'|'useLayoutEffect'|'useMemo'|'useReducer'|'useRef'|'useState')[]***REMOVED*** [expectedHookNames] React hook names to which search is limited.
     * @returns {boolean***REMOVED*** True if the node is a call to a React hook
     */
    isReactHookCall(node, expectedHookNames) {
      if (!astUtil.isCallExpression(node)) {
        return false;
      ***REMOVED***

      const defaultReactImports = components.getDefaultReactImports();
      const namedReactImports = components.getNamedReactImports();

      const defaultReactImportName = defaultReactImports
        && defaultReactImports[0]
        && defaultReactImports[0].local.name;
      const reactHookImportSpecifiers = namedReactImports
        && namedReactImports.filter((specifier) => USE_HOOK_PREFIX_REGEX.test(specifier.imported.name));
      const reactHookImportNames = reactHookImportSpecifiers
        && fromEntries(reactHookImportSpecifiers.map((specifier) => [specifier.local.name, specifier.imported.name]));

      const isPotentialReactHookCall = defaultReactImportName
        && node.callee.type === 'MemberExpression'
        && node.callee.object.type === 'Identifier'
        && node.callee.object.name === defaultReactImportName
        && node.callee.property.type === 'Identifier'
        && node.callee.property.name.match(USE_HOOK_PREFIX_REGEX);

      const isPotentialHookCall = reactHookImportNames
        && node.callee.type === 'Identifier'
        && node.callee.name.match(USE_HOOK_PREFIX_REGEX);

      const scope = (isPotentialReactHookCall || isPotentialHookCall) && getScope(context, node);

      const reactResolvedDefs = isPotentialReactHookCall
        && scope.references
        && scope.references.find(
          (reference) => reference.identifier.name === defaultReactImportName
        ).resolved.defs;

      const isReactShadowed = isPotentialReactHookCall && reactResolvedDefs
        && reactResolvedDefs.some((reactDef) => reactDef.type !== 'ImportBinding');

      const potentialHookReference = isPotentialHookCall
        && scope.references
        && scope.references.find(
          (reference) => reactHookImportNames[reference.identifier.name]
        );

      const hookResolvedDefs = potentialHookReference && potentialHookReference.resolved.defs;
      const localHookName = (
        isPotentialReactHookCall
        && node.callee.property.name
      ) || (
        isPotentialHookCall
        && potentialHookReference
        && node.callee.name
      );
      const isHookShadowed = isPotentialHookCall
        && hookResolvedDefs
        && hookResolvedDefs.some(
          (hookDef) => hookDef.name.name === localHookName
          && hookDef.type !== 'ImportBinding'
        );

      const isHookCall = (isPotentialReactHookCall && !isReactShadowed)
        || (isPotentialHookCall && localHookName && !isHookShadowed);

      if (!isHookCall) {
        return false;
      ***REMOVED***

      if (!expectedHookNames) {
        return true;
      ***REMOVED***

      return arrayIncludes(
        expectedHookNames,
        (reactHookImportNames && reactHookImportNames[localHookName]) || localHookName
      );
    ***REMOVED***,
  ***REMOVED***;

  // Component detection instructions
  const detectionInstructions = {
    CallExpression(node) {
      if (!utils.isPragmaComponentWrapper(node)) {
        return;
      ***REMOVED***
      if (node.arguments.length > 0 && astUtil.isFunctionLikeExpression(node.arguments[0])) {
        components.add(node, 2);
      ***REMOVED***
    ***REMOVED***,

    ClassExpression(node) {
      if (!componentUtil.isES6Component(node, context)) {
        return;
      ***REMOVED***
      components.add(node, 2);
    ***REMOVED***,

    ClassDeclaration(node) {
      if (!componentUtil.isES6Component(node, context)) {
        return;
      ***REMOVED***
      components.add(node, 2);
    ***REMOVED***,

    ObjectExpression(node) {
      if (!componentUtil.isES5Component(node, context)) {
        return;
      ***REMOVED***
      components.add(node, 2);
    ***REMOVED***,

    FunctionExpression(node) {
      if (node.async && node.generator) {
        components.add(node, 0);
        return;
      ***REMOVED***

      const component = utils.getStatelessComponent(node);
      if (!component) {
        return;
      ***REMOVED***
      components.add(component, 2);
    ***REMOVED***,

    FunctionDeclaration(node) {
      if (node.async && node.generator) {
        components.add(node, 0);
        return;
      ***REMOVED***

      const cNode = utils.getStatelessComponent(node);
      if (!cNode) {
        return;
      ***REMOVED***
      components.add(cNode, 2);
    ***REMOVED***,

    ArrowFunctionExpression(node) {
      const component = utils.getStatelessComponent(node);
      if (!component) {
        return;
      ***REMOVED***
      components.add(component, 2);
    ***REMOVED***,

    ThisExpression(node) {
      const component = utils.getParentStatelessComponent(node);
      if (!component || !/Function/.test(component.type) || !node.parent.property) {
        return;
      ***REMOVED***
      // Ban functions accessing a property on a ThisExpression
      components.add(node, 0);
    ***REMOVED***,
  ***REMOVED***;

  // Detect React import specifiers
  const reactImportInstructions = {
    ImportDeclaration(node) {
      const isReactImported = node.source.type === 'Literal' && node.source.value === 'react';
      if (!isReactImported) {
        return;
      ***REMOVED***

      node.specifiers.forEach((specifier) => {
        if (specifier.type === 'ImportDefaultSpecifier') {
          components.addDefaultReactImport(specifier);
        ***REMOVED***
        if (specifier.type === 'ImportSpecifier') {
          components.addNamedReactImport(specifier);
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***,
  ***REMOVED***;

  const ruleInstructions = rule(context, components, utils);
  const propTypesInstructions = propTypesUtil(context, components, utils);
  const usedPropTypesInstructions = usedPropTypesUtil(context, components, utils);
  const defaultPropsInstructions = defaultPropsUtil(context, components, utils);

  const mergedRule = mergeRules([
    detectionInstructions,
    propTypesInstructions,
    usedPropTypesInstructions,
    defaultPropsInstructions,
    reactImportInstructions,
    ruleInstructions,
  ]);

  return mergedRule;
***REMOVED***

module.exports = Object.assign(Components, {
  detect(rule) {
    return componentRule.bind(this, rule);
  ***REMOVED***,
***REMOVED***);
