// eslint-plugin-react/lib/util/componentUtil.js
'use strict';

const doctrine = require('doctrine');
const pragmaUtil = require('./pragma');
const eslintUtil = require('./eslint');

const getScope = eslintUtil.getScope;
const getSourceCode = eslintUtil.getSourceCode;
const getText = eslintUtil.getText;

// eslint-disable-next-line valid-jsdoc
/**
 * @template {(_: object) => any***REMOVED*** T
 * @param {T***REMOVED*** fn
 * @returns {T***REMOVED***
 */
function memoize(fn) {
  const cache = new WeakMap();
  // @ts-ignore
  return function memoizedFn(arg) {
    const cachedValue = cache.get(arg);
    if (cachedValue !== undefined) {
      return cachedValue;
    ***REMOVED***
    const v = fn(arg);
    cache.set(arg, v);
    return v;
  ***REMOVED***;
***REMOVED***

const getPragma = memoize(pragmaUtil.getFromContext);
const getCreateClass = memoize(pragmaUtil.getCreateClassFromContext);

/**
 * @param {ASTNode***REMOVED*** node
 * @param {Context***REMOVED*** context
 * @returns {boolean***REMOVED***
 */
function isES5Component(node, context) {
  const pragma = getPragma(context);
  const createClass = getCreateClass(context);

  if (!node.parent || !node.parent.callee) {
    return false;
  ***REMOVED***
  const callee = node.parent.callee;
  // React.createClass({***REMOVED***)
  if (callee.type === 'MemberExpression') {
    return callee.object.name === pragma && callee.property.name === createClass;
  ***REMOVED***
  // createClass({***REMOVED***)
  if (callee.type === 'Identifier') {
    return callee.name === createClass;
  ***REMOVED***
  return false;
***REMOVED***

/**
 * Check if the node is explicitly declared as a descendant of a React Component
 * @param {any***REMOVED*** node
 * @param {Context***REMOVED*** context
 * @returns {boolean***REMOVED***
 */
function isExplicitComponent(node, context) {
  const sourceCode = getSourceCode(context);
  let comment;
  // Sometimes the passed node may not have been parsed yet by eslint, and this function call crashes.
  // Can be removed when eslint sets "parent" property for all nodes on initial AST traversal: https://github.com/eslint/eslint-scope/issues/27
  // eslint-disable-next-line no-warning-comments
  // FIXME: Remove try/catch when https://github.com/eslint/eslint-scope/issues/27 is implemented.
  try {
    comment = sourceCode.getJSDocComment(node);
  ***REMOVED*** catch (e) {
    comment = null;
  ***REMOVED***

  if (comment === null) {
    return false;
  ***REMOVED***

  let commentAst;
  try {
    commentAst = doctrine.parse(comment.value, {
      unwrap: true,
      tags: ['extends', 'augments'],
    ***REMOVED***);
  ***REMOVED*** catch (e) {
    // handle a bug in the archived `doctrine`, see #2596
    return false;
  ***REMOVED***

  const relevantTags = commentAst.tags.filter((tag) => tag.name === 'React.Component' || tag.name === 'React.PureComponent');

  return relevantTags.length > 0;
***REMOVED***

/**
 * @param {ASTNode***REMOVED*** node
 * @param {Context***REMOVED*** context
 * @returns {boolean***REMOVED***
 */
function isES6Component(node, context) {
  const pragma = getPragma(context);
  if (isExplicitComponent(node, context)) {
    return true;
  ***REMOVED***

  if (!node.superClass) {
    return false;
  ***REMOVED***
  if (node.superClass.type === 'MemberExpression') {
    return node.superClass.object.name === pragma
          && /^(Pure)?Component$/.test(node.superClass.property.name);
  ***REMOVED***
  if (node.superClass.type === 'Identifier') {
    return /^(Pure)?Component$/.test(node.superClass.name);
  ***REMOVED***
  return false;
***REMOVED***

/**
 * Get the parent ES5 component node from the current scope
 * @param {Context***REMOVED*** context
 * @param {ASTNode***REMOVED*** node
 * @returns {ASTNode|null***REMOVED***
 */
function getParentES5Component(context, node) {
  let scope = getScope(context, node);
  while (scope) {
    // @ts-ignore
    node = scope.block && scope.block.parent && scope.block.parent.parent;
    if (node && isES5Component(node, context)) {
      return node;
    ***REMOVED***
    scope = scope.upper;
  ***REMOVED***
  return null;
***REMOVED***

/**
 * Get the parent ES6 component node from the current scope
 * @param {Context***REMOVED*** context
 * @param {ASTNode***REMOVED*** node
 * @returns {ASTNode | null***REMOVED***
 */
function getParentES6Component(context, node) {
  let scope = getScope(context, node);
  while (scope && scope.type !== 'class') {
    scope = scope.upper;
  ***REMOVED***
  node = scope && scope.block;
  if (!node || !isES6Component(node, context)) {
    return null;
  ***REMOVED***
  return node;
***REMOVED***

/**
 * Checks if a component extends React.PureComponent
 * @param {ASTNode***REMOVED*** node
 * @param {Context***REMOVED*** context
 * @returns {boolean***REMOVED***
 */
function isPureComponent(node, context) {
  const pragma = getPragma(context);
  if (node.superClass) {
    return new RegExp(`^(${pragma***REMOVED***\\.)?PureComponent$`).test(getText(context, node.superClass));
  ***REMOVED***
  return false;
***REMOVED***

/**
 * @param {ASTNode***REMOVED*** node
 * @returns {boolean***REMOVED***
 */
function isStateMemberExpression(node) {
  return node.type === 'MemberExpression'
    && node.object.type === 'ThisExpression'
    && node.property.name === 'state';
***REMOVED***

module.exports = {
  isES5Component,
  isES6Component,
  getParentES5Component,
  getParentES6Component,
  isExplicitComponent,
  isPureComponent,
  isStateMemberExpression,
***REMOVED***;
