// eslint-plugin-react/lib/util/props.js
/**
 * @fileoverview Utility functions for props
 */

'use strict';

const astUtil = require('./ast');

/**
 * Checks if the Identifier node passed in looks like a propTypes declaration.
 * @param {ASTNode***REMOVED*** node The node to check. Must be an Identifier node.
 * @returns {boolean***REMOVED*** `true` if the node is a propTypes declaration, `false` if not
 */
function isPropTypesDeclaration(node) {
  if (node && (node.type === 'ClassProperty' || node.type === 'PropertyDefinition')) {
    // Flow support
    if (node.typeAnnotation && node.key.name === 'props') {
      return true;
    ***REMOVED***
  ***REMOVED***
  return astUtil.getPropertyName(node) === 'propTypes';
***REMOVED***

/**
 * Checks if the node passed in looks like a contextTypes declaration.
 * @param {ASTNode***REMOVED*** node The node to check.
 * @returns {boolean***REMOVED*** `true` if the node is a contextTypes declaration, `false` if not
 */
function isContextTypesDeclaration(node) {
  if (node && (node.type === 'ClassProperty' || node.type === 'PropertyDefinition')) {
    // Flow support
    if (node.typeAnnotation && node.key.name === 'context') {
      return true;
    ***REMOVED***
  ***REMOVED***
  return astUtil.getPropertyName(node) === 'contextTypes';
***REMOVED***

/**
 * Checks if the node passed in looks like a contextType declaration.
 * @param {ASTNode***REMOVED*** node The node to check.
 * @returns {boolean***REMOVED*** `true` if the node is a contextType declaration, `false` if not
 */
function isContextTypeDeclaration(node) {
  return astUtil.getPropertyName(node) === 'contextType';
***REMOVED***

/**
 * Checks if the node passed in looks like a childContextTypes declaration.
 * @param {ASTNode***REMOVED*** node The node to check.
 * @returns {boolean***REMOVED*** `true` if the node is a childContextTypes declaration, `false` if not
 */
function isChildContextTypesDeclaration(node) {
  return astUtil.getPropertyName(node) === 'childContextTypes';
***REMOVED***

/**
 * Checks if the Identifier node passed in looks like a defaultProps declaration.
 * @param {ASTNode***REMOVED*** node The node to check. Must be an Identifier node.
 * @returns {boolean***REMOVED*** `true` if the node is a defaultProps declaration, `false` if not
 */
function isDefaultPropsDeclaration(node) {
  const propName = astUtil.getPropertyName(node);
  return (propName === 'defaultProps' || propName === 'getDefaultProps');
***REMOVED***

/**
 * Checks if we are declaring a display name
 * @param {ASTNode***REMOVED*** node The AST node being checked.
 * @returns {boolean***REMOVED*** True if we are declaring a display name, false if not.
 */
function isDisplayNameDeclaration(node) {
  switch (node.type) {
    case 'ClassProperty':
    case 'PropertyDefinition':
      return node.key && node.key.name === 'displayName';
    case 'Identifier':
      return node.name === 'displayName';
    case 'Literal':
      return node.value === 'displayName';
    default:
      return false;
  ***REMOVED***
***REMOVED***

/**
 * Checks if the PropTypes MemberExpression node passed in declares a required propType.
 * @param {ASTNode***REMOVED*** propTypeExpression node to check. Must be a `PropTypes` MemberExpression.
 * @returns {boolean***REMOVED*** `true` if this PropType is required, `false` if not.
 */
function isRequiredPropType(propTypeExpression) {
  return propTypeExpression.type === 'MemberExpression'
    && propTypeExpression.property.name === 'isRequired';
***REMOVED***

/**
 * Returns the type arguments of a node or type parameters if type arguments are not available.
 * @param {ASTNode***REMOVED*** node The node to get the type arguments from.
 * @returns {ASTNode***REMOVED*** The type arguments or type parameters of the node.
 */
function getTypeArguments(node) {
  if ('typeArguments' in node) {
    return node.typeArguments;
  ***REMOVED***
  return node.typeParameters;
***REMOVED***

/**
 * Returns the super type arguments of a node or super type parameters if type arguments are not available.
 * @param {ASTNode***REMOVED*** node The node to get the super type arguments from.
 * @returns {ASTNode***REMOVED*** The super type arguments or parameters of the node.
 */
function getSuperTypeArguments(node) {
  if ('superTypeArguments' in node) {
    return node.superTypeArguments;
  ***REMOVED***
  return node.superTypeParameters;
***REMOVED***

module.exports = {
  isPropTypesDeclaration,
  isContextTypesDeclaration,
  isContextTypeDeclaration,
  isChildContextTypesDeclaration,
  isDefaultPropsDeclaration,
  isDisplayNameDeclaration,
  isRequiredPropType,
  getTypeArguments,
  getSuperTypeArguments,
***REMOVED***;
