// eslint-plugin-react/lib/util/propWrapper.js
/**
 * @fileoverview Utility functions for propWrapperFunctions setting
 */

'use strict';

const filter = require('es-iterator-helpers/Iterator.prototype.filter');
const some = require('es-iterator-helpers/Iterator.prototype.some');

function searchPropWrapperFunctions(name, propWrapperFunctions) {
  const splitName = name.split('.');
  return some(propWrapperFunctions.values(), (func) => {
    if (splitName.length === 2 && func.object === splitName[0] && func.property === splitName[1]) {
      return true;
    ***REMOVED***
    return name === func || func.property === name;
  ***REMOVED***);
***REMOVED***

function getPropWrapperFunctions(context) {
  return new Set(context.settings.propWrapperFunctions || []);
***REMOVED***

function isPropWrapperFunction(context, name) {
  if (typeof name !== 'string') {
    return false;
  ***REMOVED***
  const propWrapperFunctions = getPropWrapperFunctions(context);
  return searchPropWrapperFunctions(name, propWrapperFunctions);
***REMOVED***

function getExactPropWrapperFunctions(context) {
  const propWrapperFunctions = getPropWrapperFunctions(context);
  const exactPropWrappers = filter(propWrapperFunctions.values(), (func) => func.exact === true);
  return new Set(exactPropWrappers);
***REMOVED***

function isExactPropWrapperFunction(context, name) {
  const exactPropWrappers = getExactPropWrapperFunctions(context);
  return searchPropWrapperFunctions(name, exactPropWrappers);
***REMOVED***

function formatPropWrapperFunctions(propWrapperFunctions) {
  return Array.from(propWrapperFunctions, (func) => {
    if (func.object && func.property) {
      return `'${func.object***REMOVED***.${func.property***REMOVED***'`;
    ***REMOVED***
    if (func.property) {
      return `'${func.property***REMOVED***'`;
    ***REMOVED***
    return `'${func***REMOVED***'`;
  ***REMOVED***).join(', ');
***REMOVED***

module.exports = {
  formatPropWrapperFunctions,
  getExactPropWrapperFunctions,
  getPropWrapperFunctions,
  isExactPropWrapperFunction,
  isPropWrapperFunction,
***REMOVED***;
