// async/internal/Heap.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
***REMOVED***);
// Binary min-heap implementation used for priority queue.
// Implementation is stable, i.e. push time is considered for equal priorities
class Heap {
    constructor() {
        this.heap = [];
        this.pushCount = Number.MIN_SAFE_INTEGER;
    ***REMOVED***

    get length() {
        return this.heap.length;
    ***REMOVED***

    empty() {
        this.heap = [];
        return this;
    ***REMOVED***

    percUp(index) {
        let p;

        while (index > 0 && smaller(this.heap[index], this.heap[p = parent(index)])) {
            let t = this.heap[index];
            this.heap[index] = this.heap[p];
            this.heap[p] = t;

            index = p;
        ***REMOVED***
    ***REMOVED***

    percDown(index) {
        let l;

        while ((l = leftChi(index)) < this.heap.length) {
            if (l + 1 < this.heap.length && smaller(this.heap[l + 1], this.heap[l])) {
                l = l + 1;
            ***REMOVED***

            if (smaller(this.heap[index], this.heap[l])) {
                break;
            ***REMOVED***

            let t = this.heap[index];
            this.heap[index] = this.heap[l];
            this.heap[l] = t;

            index = l;
        ***REMOVED***
    ***REMOVED***

    push(node) {
        node.pushCount = ++this.pushCount;
        this.heap.push(node);
        this.percUp(this.heap.length - 1);
    ***REMOVED***

    unshift(node) {
        return this.heap.push(node);
    ***REMOVED***

    shift() {
        let [top] = this.heap;

        this.heap[0] = this.heap[this.heap.length - 1];
        this.heap.pop();
        this.percDown(0);

        return top;
    ***REMOVED***

    toArray() {
        return [...this];
    ***REMOVED***

    *[Symbol.iterator]() {
        for (let i = 0; i < this.heap.length; i++) {
            yield this.heap[i].data;
        ***REMOVED***
    ***REMOVED***

    remove(testFn) {
        let j = 0;
        for (let i = 0; i < this.heap.length; i++) {
            if (!testFn(this.heap[i])) {
                this.heap[j] = this.heap[i];
                j++;
            ***REMOVED***
        ***REMOVED***

        this.heap.splice(j);

        for (let i = parent(this.heap.length - 1); i >= 0; i--) {
            this.percDown(i);
        ***REMOVED***

        return this;
    ***REMOVED***
***REMOVED***

exports.default = Heap;
function leftChi(i) {
    return (i << 1) + 1;
***REMOVED***

function parent(i) {
    return (i + 1 >> 1) - 1;
***REMOVED***

function smaller(x, y) {
    if (x.priority !== y.priority) {
        return x.priority < y.priority;
    ***REMOVED*** else {
        return x.pushCount < y.pushCount;
    ***REMOVED***
***REMOVED***
module.exports = exports.default;
