// async/internal/queue.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
***REMOVED***);
exports.default = queue;

var _onlyOnce = require('./onlyOnce.js');

var _onlyOnce2 = _interopRequireDefault(_onlyOnce);

var _setImmediate = require('./setImmediate.js');

var _setImmediate2 = _interopRequireDefault(_setImmediate);

var _DoublyLinkedList = require('./DoublyLinkedList.js');

var _DoublyLinkedList2 = _interopRequireDefault(_DoublyLinkedList);

var _wrapAsync = require('./wrapAsync.js');

var _wrapAsync2 = _interopRequireDefault(_wrapAsync);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj ***REMOVED***; ***REMOVED***

function queue(worker, concurrency, payload) {
    if (concurrency == null) {
        concurrency = 1;
    ***REMOVED*** else if (concurrency === 0) {
        throw new RangeError('Concurrency must not be zero');
    ***REMOVED***

    var _worker = (0, _wrapAsync2.default)(worker);
    var numRunning = 0;
    var workersList = [];
    const events = {
        error: [],
        drain: [],
        saturated: [],
        unsaturated: [],
        empty: []
    ***REMOVED***;

    function on(event, handler) {
        events[event].push(handler);
    ***REMOVED***

    function once(event, handler) {
        const handleAndRemove = (...args) => {
            off(event, handleAndRemove);
            handler(...args);
        ***REMOVED***;
        events[event].push(handleAndRemove);
    ***REMOVED***

    function off(event, handler) {
        if (!event) return Object.keys(events).forEach(ev => events[ev] = []);
        if (!handler) return events[event] = [];
        events[event] = events[event].filter(ev => ev !== handler);
    ***REMOVED***

    function trigger(event, ...args) {
        events[event].forEach(handler => handler(...args));
    ***REMOVED***

    var processingScheduled = false;
    function _insert(data, insertAtFront, rejectOnError, callback) {
        if (callback != null && typeof callback !== 'function') {
            throw new Error('task callback must be a function');
        ***REMOVED***
        q.started = true;

        var res, rej;
        function promiseCallback(err, ...args) {
            // we don't care about the error, let the global error handler
            // deal with it
            if (err) return rejectOnError ? rej(err) : res();
            if (args.length <= 1) return res(args[0]);
            res(args);
        ***REMOVED***

        var item = q._createTaskItem(data, rejectOnError ? promiseCallback : callback || promiseCallback);

        if (insertAtFront) {
            q._tasks.unshift(item);
        ***REMOVED*** else {
            q._tasks.push(item);
        ***REMOVED***

        if (!processingScheduled) {
            processingScheduled = true;
            (0, _setImmediate2.default)(() => {
                processingScheduled = false;
                q.process();
            ***REMOVED***);
        ***REMOVED***

        if (rejectOnError || !callback) {
            return new Promise((resolve, reject) => {
                res = resolve;
                rej = reject;
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***

    function _createCB(tasks) {
        return function (err, ...args) {
            numRunning -= 1;

            for (var i = 0, l = tasks.length; i < l; i++) {
                var task = tasks[i];

                var index = workersList.indexOf(task);
                if (index === 0) {
                    workersList.shift();
                ***REMOVED*** else if (index > 0) {
                    workersList.splice(index, 1);
                ***REMOVED***

                task.callback(err, ...args);

                if (err != null) {
                    trigger('error', err, task.data);
                ***REMOVED***
            ***REMOVED***

            if (numRunning <= q.concurrency - q.buffer) {
                trigger('unsaturated');
            ***REMOVED***

            if (q.idle()) {
                trigger('drain');
            ***REMOVED***
            q.process();
        ***REMOVED***;
    ***REMOVED***

    function _maybeDrain(data) {
        if (data.length === 0 && q.idle()) {
            // call drain immediately if there are no tasks
            (0, _setImmediate2.default)(() => trigger('drain'));
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***

    const eventMethod = name => handler => {
        if (!handler) {
            return new Promise((resolve, reject) => {
                once(name, (err, data) => {
                    if (err) return reject(err);
                    resolve(data);
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***
        off(name);
        on(name, handler);
    ***REMOVED***;

    var isProcessing = false;
    var q = {
        _tasks: new _DoublyLinkedList2.default(),
        _createTaskItem(data, callback) {
            return {
                data,
                callback
            ***REMOVED***;
        ***REMOVED***,
        *[Symbol.iterator]() {
            yield* q._tasks[Symbol.iterator]();
        ***REMOVED***,
        concurrency,
        payload,
        buffer: concurrency / 4,
        started: false,
        paused: false,
        push(data, callback) {
            if (Array.isArray(data)) {
                if (_maybeDrain(data)) return;
                return data.map(datum => _insert(datum, false, false, callback));
            ***REMOVED***
            return _insert(data, false, false, callback);
        ***REMOVED***,
        pushAsync(data, callback) {
            if (Array.isArray(data)) {
                if (_maybeDrain(data)) return;
                return data.map(datum => _insert(datum, false, true, callback));
            ***REMOVED***
            return _insert(data, false, true, callback);
        ***REMOVED***,
        kill() {
            off();
            q._tasks.empty();
        ***REMOVED***,
        unshift(data, callback) {
            if (Array.isArray(data)) {
                if (_maybeDrain(data)) return;
                return data.map(datum => _insert(datum, true, false, callback));
            ***REMOVED***
            return _insert(data, true, false, callback);
        ***REMOVED***,
        unshiftAsync(data, callback) {
            if (Array.isArray(data)) {
                if (_maybeDrain(data)) return;
                return data.map(datum => _insert(datum, true, true, callback));
            ***REMOVED***
            return _insert(data, true, true, callback);
        ***REMOVED***,
        remove(testFn) {
            q._tasks.remove(testFn);
        ***REMOVED***,
        process() {
            // Avoid trying to start too many processing operations. This can occur
            // when callbacks resolve synchronously (#1267).
            if (isProcessing) {
                return;
            ***REMOVED***
            isProcessing = true;
            while (!q.paused && numRunning < q.concurrency && q._tasks.length) {
                var tasks = [],
                    data = [];
                var l = q._tasks.length;
                if (q.payload) l = Math.min(l, q.payload);
                for (var i = 0; i < l; i++) {
                    var node = q._tasks.shift();
                    tasks.push(node);
                    workersList.push(node);
                    data.push(node.data);
                ***REMOVED***

                numRunning += 1;

                if (q._tasks.length === 0) {
                    trigger('empty');
                ***REMOVED***

                if (numRunning === q.concurrency) {
                    trigger('saturated');
                ***REMOVED***

                var cb = (0, _onlyOnce2.default)(_createCB(tasks));
                _worker(data, cb);
            ***REMOVED***
            isProcessing = false;
        ***REMOVED***,
        length() {
            return q._tasks.length;
        ***REMOVED***,
        running() {
            return numRunning;
        ***REMOVED***,
        workersList() {
            return workersList;
        ***REMOVED***,
        idle() {
            return q._tasks.length + numRunning === 0;
        ***REMOVED***,
        pause() {
            q.paused = true;
        ***REMOVED***,
        resume() {
            if (q.paused === false) {
                return;
            ***REMOVED***
            q.paused = false;
            (0, _setImmediate2.default)(q.process);
        ***REMOVED***
    ***REMOVED***;
    // define these as fixed properties, so people get useful errors when updating
    Object.defineProperties(q, {
        saturated: {
            writable: false,
            value: eventMethod('saturated')
        ***REMOVED***,
        unsaturated: {
            writable: false,
            value: eventMethod('unsaturated')
        ***REMOVED***,
        empty: {
            writable: false,
            value: eventMethod('empty')
        ***REMOVED***,
        drain: {
            writable: false,
            value: eventMethod('drain')
        ***REMOVED***,
        error: {
            writable: false,
            value: eventMethod('error')
        ***REMOVED***
    ***REMOVED***);
    return q;
***REMOVED***
module.exports = exports.default;
