// eslint-plugin-testing-library/rules/consistent-data-testid.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.RULE_NAME = void 0;
const create_testing_library_rule_1 = require("../create-testing-library-rule");
const node_utils_1 = require("../node-utils");
exports.RULE_NAME = 'consistent-data-testid';
const FILENAME_PLACEHOLDER = '{fileName***REMOVED***';
exports.default = (0, create_testing_library_rule_1.createTestingLibraryRule)({
    name: exports.RULE_NAME,
    meta: {
        type: 'suggestion',
        docs: {
            description: 'Ensures consistent usage of `data-testid`',
            recommendedConfig: {
                dom: false,
                angular: false,
                react: false,
                vue: false,
                marko: false,
            ***REMOVED***,
        ***REMOVED***,
        messages: {
            consistentDataTestId: '`{{attr***REMOVED******REMOVED***` "{{value***REMOVED******REMOVED***" should match `{{regex***REMOVED******REMOVED***`',
            consistentDataTestIdCustomMessage: '`{{message***REMOVED******REMOVED***`',
        ***REMOVED***,
        schema: [
            {
                type: 'object',
                default: {***REMOVED***,
                additionalProperties: false,
                required: ['testIdPattern'],
                properties: {
                    testIdPattern: {
                        type: 'string',
                    ***REMOVED***,
                    testIdAttribute: {
                        default: 'data-testid',
                        oneOf: [
                            {
                                type: 'string',
                            ***REMOVED***,
                            {
                                type: 'array',
                                items: {
                                    type: 'string',
                                ***REMOVED***,
                            ***REMOVED***,
                        ],
                    ***REMOVED***,
                    customMessage: {
                        default: undefined,
                        type: 'string',
                    ***REMOVED***,
                ***REMOVED***,
            ***REMOVED***,
        ],
    ***REMOVED***,
    defaultOptions: [
        {
            testIdPattern: '',
            testIdAttribute: 'data-testid',
            customMessage: undefined,
        ***REMOVED***,
    ],
    detectionOptions: {
        skipRuleReportingCheck: true,
    ***REMOVED***,
    create: (context, [options]) => {
        const { getFilename ***REMOVED*** = context;
        const { testIdPattern, testIdAttribute: attr, customMessage ***REMOVED*** = options;
        function getFileNameData() {
            var _a;
            const splitPath = getFilename().split('/');
            const fileNameWithExtension = (_a = splitPath.pop()) !== null && _a !== void 0 ? _a : '';
            if (fileNameWithExtension.includes('[') ||
                fileNameWithExtension.includes(']')) {
                return { fileName: undefined ***REMOVED***;
            ***REMOVED***
            const parent = splitPath.pop();
            const fileName = fileNameWithExtension.split('.').shift();
            return {
                fileName: fileName === 'index' ? parent : fileName,
            ***REMOVED***;
        ***REMOVED***
        function getTestIdValidator(fileName) {
            return new RegExp(testIdPattern.replace(FILENAME_PLACEHOLDER, fileName));
        ***REMOVED***
        function isTestIdAttribute(name) {
            var _a;
            if (typeof attr === 'string') {
                return attr === name;
            ***REMOVED***
            else {
                return (_a = attr === null || attr === void 0 ? void 0 : attr.includes(name)) !== null && _a !== void 0 ? _a : false;
            ***REMOVED***
        ***REMOVED***
        function getErrorMessageId() {
            if (customMessage === undefined) {
                return 'consistentDataTestId';
            ***REMOVED***
            return 'consistentDataTestIdCustomMessage';
        ***REMOVED***
        return {
            JSXIdentifier: (node) => {
                if (!node.parent ||
                    !(0, node_utils_1.isJSXAttribute)(node.parent) ||
                    !(0, node_utils_1.isLiteral)(node.parent.value) ||
                    !isTestIdAttribute(node.name)) {
                    return;
                ***REMOVED***
                const value = node.parent.value.value;
                const { fileName ***REMOVED*** = getFileNameData();
                const regex = getTestIdValidator(fileName !== null && fileName !== void 0 ? fileName : '');
                if (value && typeof value === 'string' && !regex.test(value)) {
                    context.report({
                        node,
                        messageId: getErrorMessageId(),
                        data: {
                            attr: node.name,
                            value,
                            regex,
                            message: customMessage,
                        ***REMOVED***,
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***;
    ***REMOVED***,
***REMOVED***);
