// eslint-plugin-testing-library/rules/no-node-access.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.RULE_NAME = void 0;
const utils_1 = require("@typescript-eslint/utils");
const create_testing_library_rule_1 = require("../create-testing-library-rule");
const utils_2 = require("../utils");
exports.RULE_NAME = 'no-node-access';
exports.default = (0, create_testing_library_rule_1.createTestingLibraryRule)({
    name: exports.RULE_NAME,
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallow direct Node access',
            recommendedConfig: {
                dom: false,
                angular: 'error',
                react: 'error',
                vue: 'error',
                marko: 'error',
            ***REMOVED***,
        ***REMOVED***,
        messages: {
            noNodeAccess: 'Avoid direct Node access. Prefer using the methods from Testing Library.',
        ***REMOVED***,
        schema: [
            {
                type: 'object',
                properties: {
                    allowContainerFirstChild: {
                        type: 'boolean',
                    ***REMOVED***,
                ***REMOVED***,
            ***REMOVED***,
        ],
    ***REMOVED***,
    defaultOptions: [
        {
            allowContainerFirstChild: false,
        ***REMOVED***,
    ],
    create(context, [{ allowContainerFirstChild = false ***REMOVED***], helpers) {
        function showErrorForNodeAccess(node) {
            if (!helpers.isTestingLibraryImported(true)) {
                return;
            ***REMOVED***
            if (utils_1.ASTUtils.isIdentifier(node.property) &&
                utils_2.ALL_RETURNING_NODES.includes(node.property.name)) {
                if (allowContainerFirstChild && node.property.name === 'firstChild') {
                    return;
                ***REMOVED***
                if (utils_1.ASTUtils.isIdentifier(node.object) &&
                    node.object.name === 'props') {
                    return;
                ***REMOVED***
                context.report({
                    node,
                    loc: node.property.loc.start,
                    messageId: 'noNodeAccess',
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***
        return {
            'ExpressionStatement MemberExpression': showErrorForNodeAccess,
            'VariableDeclarator MemberExpression': showErrorForNodeAccess,
        ***REMOVED***;
    ***REMOVED***,
***REMOVED***);
