// eslint-plugin-testing-library/rules/no-render-in-setup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.findClosestBeforeHook = exports.RULE_NAME = void 0;
const utils_1 = require("@typescript-eslint/utils");
const create_testing_library_rule_1 = require("../create-testing-library-rule");
const node_utils_1 = require("../node-utils");
const utils_2 = require("../utils");
exports.RULE_NAME = 'no-render-in-setup';
function findClosestBeforeHook(node, testingFrameworkSetupHooksToFilter) {
    if (node === null) {
        return null;
    ***REMOVED***
    if ((0, node_utils_1.isCallExpression)(node) &&
        utils_1.ASTUtils.isIdentifier(node.callee) &&
        testingFrameworkSetupHooksToFilter.includes(node.callee.name)) {
        return node.callee;
    ***REMOVED***
    if (node.parent) {
        return findClosestBeforeHook(node.parent, testingFrameworkSetupHooksToFilter);
    ***REMOVED***
    return null;
***REMOVED***
exports.findClosestBeforeHook = findClosestBeforeHook;
exports.default = (0, create_testing_library_rule_1.createTestingLibraryRule)({
    name: exports.RULE_NAME,
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallow the use of `render` in testing frameworks setup functions',
            recommendedConfig: {
                dom: false,
                angular: 'error',
                react: 'error',
                vue: 'error',
                marko: 'error',
            ***REMOVED***,
        ***REMOVED***,
        messages: {
            noRenderInSetup: 'Forbidden usage of `render` within testing framework `{{ name ***REMOVED******REMOVED***` setup',
        ***REMOVED***,
        schema: [
            {
                type: 'object',
                properties: {
                    allowTestingFrameworkSetupHook: {
                        enum: utils_2.TESTING_FRAMEWORK_SETUP_HOOKS,
                    ***REMOVED***,
                ***REMOVED***,
            ***REMOVED***,
        ],
    ***REMOVED***,
    defaultOptions: [
        {
            allowTestingFrameworkSetupHook: '',
        ***REMOVED***,
    ],
    create(context, [{ allowTestingFrameworkSetupHook ***REMOVED***], helpers) {
        const renderWrapperNames = [];
        function detectRenderWrapper(node) {
            const innerFunction = (0, node_utils_1.getInnermostReturningFunction)(context, node);
            if (innerFunction) {
                renderWrapperNames.push((0, node_utils_1.getFunctionName)(innerFunction));
            ***REMOVED***
        ***REMOVED***
        return {
            CallExpression(node) {
                const testingFrameworkSetupHooksToFilter = utils_2.TESTING_FRAMEWORK_SETUP_HOOKS.filter((hook) => hook !== allowTestingFrameworkSetupHook);
                const callExpressionIdentifier = (0, node_utils_1.getDeepestIdentifierNode)(node);
                if (!callExpressionIdentifier) {
                    return;
                ***REMOVED***
                const isRenderIdentifier = helpers.isRenderUtil(callExpressionIdentifier);
                if (isRenderIdentifier) {
                    detectRenderWrapper(callExpressionIdentifier);
                ***REMOVED***
                if (!isRenderIdentifier &&
                    !renderWrapperNames.includes(callExpressionIdentifier.name)) {
                    return;
                ***REMOVED***
                const beforeHook = findClosestBeforeHook(node, testingFrameworkSetupHooksToFilter);
                if (!beforeHook) {
                    return;
                ***REMOVED***
                context.report({
                    node: callExpressionIdentifier,
                    messageId: 'noRenderInSetup',
                    data: {
                        name: beforeHook.name,
                    ***REMOVED***,
                ***REMOVED***);
            ***REMOVED***,
        ***REMOVED***;
    ***REMOVED***,
***REMOVED***);
