// eslint-plugin-testing-library/rules/no-wait-for-multiple-assertions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.RULE_NAME = void 0;
const create_testing_library_rule_1 = require("../create-testing-library-rule");
const node_utils_1 = require("../node-utils");
exports.RULE_NAME = 'no-wait-for-multiple-assertions';
exports.default = (0, create_testing_library_rule_1.createTestingLibraryRule)({
    name: exports.RULE_NAME,
    meta: {
        type: 'suggestion',
        docs: {
            description: 'Disallow the use of multiple `expect` calls inside `waitFor`',
            recommendedConfig: {
                dom: 'error',
                angular: 'error',
                react: 'error',
                vue: 'error',
                marko: 'error',
            ***REMOVED***,
        ***REMOVED***,
        messages: {
            noWaitForMultipleAssertion: 'Avoid using multiple assertions within `waitFor` callback',
        ***REMOVED***,
        schema: [],
    ***REMOVED***,
    defaultOptions: [],
    create(context, _, helpers) {
        function getExpectNodes(body) {
            return body.filter((node) => {
                if (!(0, node_utils_1.isExpressionStatement)(node)) {
                    return false;
                ***REMOVED***
                const expressionIdentifier = (0, node_utils_1.getPropertyIdentifierNode)(node);
                if (!expressionIdentifier) {
                    return false;
                ***REMOVED***
                return expressionIdentifier.name === 'expect';
            ***REMOVED***);
        ***REMOVED***
        function reportMultipleAssertion(node) {
            if (!node.parent) {
                return;
            ***REMOVED***
            const callExpressionNode = node.parent.parent;
            const callExpressionIdentifier = (0, node_utils_1.getPropertyIdentifierNode)(callExpressionNode);
            if (!callExpressionIdentifier) {
                return;
            ***REMOVED***
            if (!helpers.isAsyncUtil(callExpressionIdentifier, ['waitFor'])) {
                return;
            ***REMOVED***
            const expectNodes = getExpectNodes(node.body);
            if (expectNodes.length <= 1) {
                return;
            ***REMOVED***
            for (let i = 0; i < expectNodes.length; i++) {
                if (i !== 0) {
                    context.report({
                        node: expectNodes[i],
                        messageId: 'noWaitForMultipleAssertion',
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        return {
            'CallExpression > ArrowFunctionExpression > BlockStatement': reportMultipleAssertion,
            'CallExpression > FunctionExpression > BlockStatement': reportMultipleAssertion,
        ***REMOVED***;
    ***REMOVED***,
***REMOVED***);
