// eslint-plugin-testing-library/rules/prefer-query-matchers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.RULE_NAME = void 0;
const create_testing_library_rule_1 = require("../create-testing-library-rule");
const node_utils_1 = require("../node-utils");
exports.RULE_NAME = 'prefer-query-matchers';
exports.default = (0, create_testing_library_rule_1.createTestingLibraryRule)({
    name: exports.RULE_NAME,
    meta: {
        docs: {
            description: 'Ensure the configured `get*`/`query*` query is used with the corresponding matchers',
            recommendedConfig: {
                dom: false,
                angular: false,
                react: false,
                vue: false,
                marko: false,
            ***REMOVED***,
        ***REMOVED***,
        messages: {
            wrongQueryForMatcher: 'Use `{{ query ***REMOVED******REMOVED***By*` queries for {{ matcher ***REMOVED******REMOVED***',
        ***REMOVED***,
        schema: [
            {
                type: 'object',
                additionalProperties: false,
                properties: {
                    validEntries: {
                        type: 'array',
                        items: {
                            type: 'object',
                            properties: {
                                query: {
                                    type: 'string',
                                    enum: ['get', 'query'],
                                ***REMOVED***,
                                matcher: {
                                    type: 'string',
                                ***REMOVED***,
                            ***REMOVED***,
                        ***REMOVED***,
                    ***REMOVED***,
                ***REMOVED***,
            ***REMOVED***,
        ],
        type: 'suggestion',
    ***REMOVED***,
    defaultOptions: [
        {
            validEntries: [],
        ***REMOVED***,
    ],
    create(context, [{ validEntries ***REMOVED***], helpers) {
        return {
            'CallExpression Identifier'(node) {
                const expectCallNode = (0, node_utils_1.findClosestCallNode)(node, 'expect');
                if (!expectCallNode || !(0, node_utils_1.isMemberExpression)(expectCallNode.parent)) {
                    return;
                ***REMOVED***
                if (!helpers.isSyncQuery(node)) {
                    return;
                ***REMOVED***
                const isGetBy = helpers.isGetQueryVariant(node);
                const expectStatement = expectCallNode.parent;
                for (const entry of validEntries) {
                    const { query, matcher ***REMOVED*** = entry;
                    const isMatchingAssertForThisEntry = helpers.isMatchingAssert(expectStatement, matcher);
                    if (!isMatchingAssertForThisEntry) {
                        continue;
                    ***REMOVED***
                    const actualQuery = isGetBy ? 'get' : 'query';
                    if (query !== actualQuery) {
                        context.report({
                            node,
                            messageId: 'wrongQueryForMatcher',
                            data: { query, matcher ***REMOVED***,
                        ***REMOVED***);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***;
    ***REMOVED***,
***REMOVED***);
