// eslint-plugin-testing-library/rules/prefer-presence-queries.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.RULE_NAME = void 0;
const create_testing_library_rule_1 = require("../create-testing-library-rule");
const node_utils_1 = require("../node-utils");
exports.RULE_NAME = 'prefer-presence-queries';
exports.default = (0, create_testing_library_rule_1.createTestingLibraryRule)({
    name: exports.RULE_NAME,
    meta: {
        docs: {
            description: 'Ensure appropriate `get*`/`query*` queries are used with their respective matchers',
            recommendedConfig: {
                dom: 'error',
                angular: 'error',
                react: 'error',
                vue: 'error',
                marko: 'error',
            ***REMOVED***,
        ***REMOVED***,
        messages: {
            wrongPresenceQuery: 'Use `getBy*` queries rather than `queryBy*` for checking element is present',
            wrongAbsenceQuery: 'Use `queryBy*` queries rather than `getBy*` for checking element is NOT present',
        ***REMOVED***,
        schema: [
            {
                type: 'object',
                additionalProperties: false,
                properties: {
                    presence: {
                        type: 'boolean',
                    ***REMOVED***,
                    absence: {
                        type: 'boolean',
                    ***REMOVED***,
                ***REMOVED***,
            ***REMOVED***,
        ],
        type: 'suggestion',
    ***REMOVED***,
    defaultOptions: [
        {
            presence: true,
            absence: true,
        ***REMOVED***,
    ],
    create(context, [{ absence = true, presence = true ***REMOVED***], helpers) {
        return {
            'CallExpression Identifier'(node) {
                const expectCallNode = (0, node_utils_1.findClosestCallNode)(node, 'expect');
                const withinCallNode = (0, node_utils_1.findClosestCallNode)(node, 'within');
                if (!expectCallNode || !(0, node_utils_1.isMemberExpression)(expectCallNode.parent)) {
                    return;
                ***REMOVED***
                if (!helpers.isSyncQuery(node)) {
                    return;
                ***REMOVED***
                const isPresenceQuery = helpers.isGetQueryVariant(node);
                const expectStatement = expectCallNode.parent;
                const isPresenceAssert = helpers.isPresenceAssert(expectStatement);
                const isAbsenceAssert = helpers.isAbsenceAssert(expectStatement);
                if (!isPresenceAssert && !isAbsenceAssert) {
                    return;
                ***REMOVED***
                if (presence &&
                    (withinCallNode || isPresenceAssert) &&
                    !isPresenceQuery) {
                    context.report({ node, messageId: 'wrongPresenceQuery' ***REMOVED***);
                ***REMOVED***
                else if (!withinCallNode &&
                    absence &&
                    isAbsenceAssert &&
                    isPresenceQuery) {
                    context.report({ node, messageId: 'wrongAbsenceQuery' ***REMOVED***);
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***;
    ***REMOVED***,
***REMOVED***);
