// eslint-plugin-testing-library/rules/prefer-user-event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.MAPPING_TO_USER_EVENT = exports.UserEventMethods = exports.RULE_NAME = void 0;
const utils_1 = require("@typescript-eslint/utils");
const create_testing_library_rule_1 = require("../create-testing-library-rule");
const node_utils_1 = require("../node-utils");
exports.RULE_NAME = 'prefer-user-event';
exports.UserEventMethods = [
    'click',
    'dblClick',
    'type',
    'upload',
    'clear',
    'selectOptions',
    'deselectOptions',
    'tab',
    'hover',
    'unhover',
    'paste',
];
exports.MAPPING_TO_USER_EVENT = {
    click: ['click', 'type', 'selectOptions', 'deselectOptions'],
    change: ['upload', 'type', 'clear', 'selectOptions', 'deselectOptions'],
    dblClick: ['dblClick'],
    input: ['type', 'upload', 'selectOptions', 'deselectOptions', 'paste'],
    keyDown: ['type', 'tab'],
    keyPress: ['type'],
    keyUp: ['type', 'tab'],
    mouseDown: ['click', 'dblClick', 'selectOptions', 'deselectOptions'],
    mouseEnter: ['hover', 'selectOptions', 'deselectOptions'],
    mouseLeave: ['unhover'],
    mouseMove: ['hover', 'unhover', 'selectOptions', 'deselectOptions'],
    mouseOut: ['unhover'],
    mouseOver: ['hover', 'selectOptions', 'deselectOptions'],
    mouseUp: ['click', 'dblClick', 'selectOptions', 'deselectOptions'],
    paste: ['paste'],
    pointerDown: ['click', 'dblClick', 'selectOptions', 'deselectOptions'],
    pointerEnter: ['hover', 'selectOptions', 'deselectOptions'],
    pointerLeave: ['unhover'],
    pointerMove: ['hover', 'unhover', 'selectOptions', 'deselectOptions'],
    pointerOut: ['unhover'],
    pointerOver: ['hover', 'selectOptions', 'deselectOptions'],
    pointerUp: ['click', 'dblClick', 'selectOptions', 'deselectOptions'],
***REMOVED***;
function buildErrorMessage(fireEventMethod) {
    const userEventMethods = exports.MAPPING_TO_USER_EVENT[fireEventMethod].map((methodName) => `userEvent.${methodName***REMOVED***`);
    return userEventMethods.join(', ').replace(/, ([a-zA-Z.]+)$/, ', or $1');
***REMOVED***
const fireEventMappedMethods = Object.keys(exports.MAPPING_TO_USER_EVENT);
exports.default = (0, create_testing_library_rule_1.createTestingLibraryRule)({
    name: exports.RULE_NAME,
    meta: {
        type: 'suggestion',
        docs: {
            description: 'Suggest using `userEvent` over `fireEvent` for simulating user interactions',
            recommendedConfig: {
                dom: false,
                angular: false,
                react: false,
                vue: false,
                marko: false,
            ***REMOVED***,
        ***REMOVED***,
        messages: {
            preferUserEvent: 'Prefer using {{userEventMethods***REMOVED******REMOVED*** over fireEvent.{{fireEventMethod***REMOVED******REMOVED***',
        ***REMOVED***,
        schema: [
            {
                type: 'object',
                properties: {
                    allowedMethods: { type: 'array' ***REMOVED***,
                ***REMOVED***,
            ***REMOVED***,
        ],
    ***REMOVED***,
    defaultOptions: [{ allowedMethods: [] ***REMOVED***],
    create(context, [options], helpers) {
        const { allowedMethods ***REMOVED*** = options;
        const createEventVariables = {***REMOVED***;
        const isfireEventMethodAllowed = (methodName) => !fireEventMappedMethods.includes(methodName) ||
            allowedMethods.includes(methodName);
        const getFireEventMethodName = (callExpressionNode, node) => {
            if (!utils_1.ASTUtils.isIdentifier(callExpressionNode.callee) &&
                !(0, node_utils_1.isMemberExpression)(callExpressionNode.callee)) {
                return node.name;
            ***REMOVED***
            const secondArgument = callExpressionNode.arguments[1];
            if (utils_1.ASTUtils.isIdentifier(secondArgument) &&
                createEventVariables[secondArgument.name] !== undefined) {
                return createEventVariables[secondArgument.name];
            ***REMOVED***
            if (!(0, node_utils_1.isCallExpression)(secondArgument) ||
                !helpers.isCreateEventUtil(secondArgument)) {
                return node.name;
            ***REMOVED***
            if (utils_1.ASTUtils.isIdentifier(secondArgument.callee)) {
                return secondArgument.arguments[0]
                    .value;
            ***REMOVED***
            return secondArgument.callee
                .property.name;
        ***REMOVED***;
        return {
            'CallExpression Identifier'(node) {
                if (!helpers.isFireEventMethod(node)) {
                    return;
                ***REMOVED***
                const closestCallExpression = (0, node_utils_1.findClosestCallExpressionNode)(node, true);
                if (!closestCallExpression) {
                    return;
                ***REMOVED***
                const fireEventMethodName = getFireEventMethodName(closestCallExpression, node);
                if (!fireEventMethodName ||
                    isfireEventMethodAllowed(fireEventMethodName)) {
                    return;
                ***REMOVED***
                context.report({
                    node: closestCallExpression.callee,
                    messageId: 'preferUserEvent',
                    data: {
                        userEventMethods: buildErrorMessage(fireEventMethodName),
                        fireEventMethod: fireEventMethodName,
                    ***REMOVED***,
                ***REMOVED***);
            ***REMOVED***,
            VariableDeclarator(node) {
                if (!(0, node_utils_1.isCallExpression)(node.init) ||
                    !helpers.isCreateEventUtil(node.init) ||
                    !utils_1.ASTUtils.isIdentifier(node.id)) {
                    return;
                ***REMOVED***
                let fireEventMethodName = '';
                if ((0, node_utils_1.isMemberExpression)(node.init.callee) &&
                    utils_1.ASTUtils.isIdentifier(node.init.callee.property)) {
                    fireEventMethodName = node.init.callee.property.name;
                ***REMOVED***
                else if (node.init.arguments.length > 0) {
                    fireEventMethodName = node.init.arguments[0]
                        .value;
                ***REMOVED***
                if (!isfireEventMethodAllowed(fireEventMethodName)) {
                    createEventVariables[node.id.name] = fireEventMethodName;
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***;
    ***REMOVED***,
***REMOVED***);
