// eslint-plugin-testing-library/rules/prefer-query-by-disappearance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.RULE_NAME = void 0;
const create_testing_library_rule_1 = require("../create-testing-library-rule");
const node_utils_1 = require("../node-utils");
exports.RULE_NAME = 'prefer-query-by-disappearance';
exports.default = (0, create_testing_library_rule_1.createTestingLibraryRule)({
    name: exports.RULE_NAME,
    meta: {
        type: 'problem',
        docs: {
            description: 'Suggest using `queryBy*` queries when waiting for disappearance',
            recommendedConfig: {
                dom: 'error',
                angular: 'error',
                react: 'error',
                vue: 'error',
                marko: 'error',
            ***REMOVED***,
        ***REMOVED***,
        messages: {
            preferQueryByDisappearance: 'Prefer using queryBy* when waiting for disappearance',
        ***REMOVED***,
        schema: [],
    ***REMOVED***,
    defaultOptions: [],
    create(context, _, helpers) {
        function isWaitForElementToBeRemoved(node) {
            const identifierNode = (0, node_utils_1.getPropertyIdentifierNode)(node);
            if (!identifierNode) {
                return false;
            ***REMOVED***
            return helpers.isAsyncUtil(identifierNode, ['waitForElementToBeRemoved']);
        ***REMOVED***
        function reportExpression(node) {
            const argumentProperty = (0, node_utils_1.isMemberExpression)(node)
                ? (0, node_utils_1.getPropertyIdentifierNode)(node.property)
                : (0, node_utils_1.getPropertyIdentifierNode)(node);
            if (!argumentProperty) {
                return false;
            ***REMOVED***
            if (helpers.isGetQueryVariant(argumentProperty) ||
                helpers.isFindQueryVariant(argumentProperty)) {
                context.report({
                    node: argumentProperty,
                    messageId: 'preferQueryByDisappearance',
                ***REMOVED***);
                return true;
            ***REMOVED***
            return false;
        ***REMOVED***
        function checkNonCallbackViolation(node) {
            if (!(0, node_utils_1.isCallExpression)(node)) {
                return false;
            ***REMOVED***
            if (!(0, node_utils_1.isMemberExpression)(node.callee) &&
                !(0, node_utils_1.getPropertyIdentifierNode)(node.callee)) {
                return false;
            ***REMOVED***
            return reportExpression(node.callee);
        ***REMOVED***
        function isReturnViolation(node) {
            if (!(0, node_utils_1.isReturnStatement)(node) || !(0, node_utils_1.isCallExpression)(node.argument)) {
                return false;
            ***REMOVED***
            return reportExpression(node.argument.callee);
        ***REMOVED***
        function isNonReturnViolation(node) {
            if (!(0, node_utils_1.isExpressionStatement)(node) || !(0, node_utils_1.isCallExpression)(node.expression)) {
                return false;
            ***REMOVED***
            if (!(0, node_utils_1.isMemberExpression)(node.expression.callee) &&
                !(0, node_utils_1.getPropertyIdentifierNode)(node.expression.callee)) {
                return false;
            ***REMOVED***
            return reportExpression(node.expression.callee);
        ***REMOVED***
        function isStatementViolation(statement) {
            return isReturnViolation(statement) || isNonReturnViolation(statement);
        ***REMOVED***
        function checkFunctionExpressionViolation(node) {
            if (!(0, node_utils_1.isFunctionExpression)(node)) {
                return false;
            ***REMOVED***
            return node.body.body.some((statement) => isStatementViolation(statement));
        ***REMOVED***
        function isArrowFunctionBodyViolation(node) {
            if (!(0, node_utils_1.isArrowFunctionExpression)(node) || !(0, node_utils_1.isBlockStatement)(node.body)) {
                return false;
            ***REMOVED***
            return node.body.body.some((statement) => isStatementViolation(statement));
        ***REMOVED***
        function isArrowFunctionImplicitReturnViolation(node) {
            if (!(0, node_utils_1.isArrowFunctionExpression)(node) || !(0, node_utils_1.isCallExpression)(node.body)) {
                return false;
            ***REMOVED***
            if (!(0, node_utils_1.isMemberExpression)(node.body.callee) &&
                !(0, node_utils_1.getPropertyIdentifierNode)(node.body.callee)) {
                return false;
            ***REMOVED***
            return reportExpression(node.body.callee);
        ***REMOVED***
        function checkArrowFunctionViolation(node) {
            return (isArrowFunctionBodyViolation(node) ||
                isArrowFunctionImplicitReturnViolation(node));
        ***REMOVED***
        function check(node) {
            if (!isWaitForElementToBeRemoved(node)) {
                return;
            ***REMOVED***
            const argumentNode = node.arguments[0];
            checkNonCallbackViolation(argumentNode);
            checkArrowFunctionViolation(argumentNode);
            checkFunctionExpressionViolation(argumentNode);
        ***REMOVED***
        return {
            CallExpression: check,
        ***REMOVED***;
    ***REMOVED***,
***REMOVED***);
