// react-scripts/scripts/init.js
// @remove-file-on-eject
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
'use strict';

// Makes the script crash on unhandled rejections instead of silently
// ignoring them. In the future, promise rejections that are not handled will
// terminate the Node.js process with a non-zero exit code.
process.on('unhandledRejection', err => {
  throw err;
***REMOVED***);

const fs = require('fs-extra');
const path = require('path');
const chalk = require('react-dev-utils/chalk');
const execSync = require('child_process').execSync;
const spawn = require('react-dev-utils/crossSpawn');
const { defaultBrowsers ***REMOVED*** = require('react-dev-utils/browsersHelper');
const os = require('os');
const verifyTypeScriptSetup = require('./utils/verifyTypeScriptSetup');

function isInGitRepository() {
  try {
    execSync('git rev-parse --is-inside-work-tree', { stdio: 'ignore' ***REMOVED***);
    return true;
  ***REMOVED*** catch (e) {
    return false;
  ***REMOVED***
***REMOVED***

function isInMercurialRepository() {
  try {
    execSync('hg --cwd . root', { stdio: 'ignore' ***REMOVED***);
    return true;
  ***REMOVED*** catch (e) {
    return false;
  ***REMOVED***
***REMOVED***

function tryGitInit() {
  try {
    execSync('git --version', { stdio: 'ignore' ***REMOVED***);
    if (isInGitRepository() || isInMercurialRepository()) {
      return false;
    ***REMOVED***

    execSync('git init', { stdio: 'ignore' ***REMOVED***);
    return true;
  ***REMOVED*** catch (e) {
    console.warn('Git repo not initialized', e);
    return false;
  ***REMOVED***
***REMOVED***

function tryGitCommit(appPath) {
  try {
    execSync('git add -A', { stdio: 'ignore' ***REMOVED***);
    execSync('git commit -m "Initialize project using Create React App"', {
      stdio: 'ignore',
    ***REMOVED***);
    return true;
  ***REMOVED*** catch (e) {
    // We couldn't commit in already initialized git repo,
    // maybe the commit author config is not set.
    // In the future, we might supply our own committer
    // like Ember CLI does, but for now, let's just
    // remove the Git files to avoid a half-done state.
    console.warn('Git commit not created', e);
    console.warn('Removing .git directory...');
    try {
      // unlinkSync() doesn't work on directories.
      fs.removeSync(path.join(appPath, '.git'));
    ***REMOVED*** catch (removeErr) {
      // Ignore.
    ***REMOVED***
    return false;
  ***REMOVED***
***REMOVED***

module.exports = function (
  appPath,
  appName,
  verbose,
  originalDirectory,
  templateName
) {
  const appPackage = require(path.join(appPath, 'package.json'));
  const useYarn = fs.existsSync(path.join(appPath, 'yarn.lock'));

  if (!templateName) {
    console.log('');
    console.error(
      `A template was not provided. This is likely because you're using an outdated version of ${chalk.cyan(
        'create-react-app'
      )***REMOVED***.`
    );
    console.error(
      `Please note that global installs of ${chalk.cyan(
        'create-react-app'
      )***REMOVED*** are no longer supported.`
    );
    console.error(
      `You can fix this by running ${chalk.cyan(
        'npm uninstall -g create-react-app'
      )***REMOVED*** or ${chalk.cyan(
        'yarn global remove create-react-app'
      )***REMOVED*** before using ${chalk.cyan('create-react-app')***REMOVED*** again.`
    );
    return;
  ***REMOVED***

  const templatePath = path.dirname(
    require.resolve(`${templateName***REMOVED***/package.json`, { paths: [appPath] ***REMOVED***)
  );

  const templateJsonPath = path.join(templatePath, 'template.json');

  let templateJson = {***REMOVED***;
  if (fs.existsSync(templateJsonPath)) {
    templateJson = require(templateJsonPath);
  ***REMOVED***

  const templatePackage = templateJson.package || {***REMOVED***;

  // This was deprecated in CRA v5.
  if (templateJson.dependencies || templateJson.scripts) {
    console.log();
    console.log(
      chalk.red(
        'Root-level `dependencies` and `scripts` keys in `template.json` were deprecated for Create React App 5.\n' +
          'This template needs to be updated to use the new `package` key.'
      )
    );
    console.log('For more information, visit https://cra.link/templates');
  ***REMOVED***

  // Keys to ignore in templatePackage
  const templatePackageBlacklist = [
    'name',
    'version',
    'description',
    'keywords',
    'bugs',
    'license',
    'author',
    'contributors',
    'files',
    'browser',
    'bin',
    'man',
    'directories',
    'repository',
    'peerDependencies',
    'bundledDependencies',
    'optionalDependencies',
    'engineStrict',
    'os',
    'cpu',
    'preferGlobal',
    'private',
    'publishConfig',
  ];

  // Keys from templatePackage that will be merged with appPackage
  const templatePackageToMerge = ['dependencies', 'scripts'];

  // Keys from templatePackage that will be added to appPackage,
  // replacing any existing entries.
  const templatePackageToReplace = Object.keys(templatePackage).filter(key => {
    return (
      !templatePackageBlacklist.includes(key) &&
      !templatePackageToMerge.includes(key)
    );
  ***REMOVED***);

  // Copy over some of the devDependencies
  appPackage.dependencies = appPackage.dependencies || {***REMOVED***;

  // Setup the script rules
  const templateScripts = templatePackage.scripts || {***REMOVED***;
  appPackage.scripts = Object.assign(
    {
      start: 'react-scripts start',
      build: 'react-scripts build',
      test: 'react-scripts test',
      eject: 'react-scripts eject',
    ***REMOVED***,
    templateScripts
  );

  // Update scripts for Yarn users
  if (useYarn) {
    appPackage.scripts = Object.entries(appPackage.scripts).reduce(
      (acc, [key, value]) => ({
        ...acc,
        [key]: value.replace(/(npm run |npm )/, 'yarn '),
      ***REMOVED***),
      {***REMOVED***
    );
  ***REMOVED***

  // Setup the eslint config
  appPackage.eslintConfig = {
    extends: 'react-app',
  ***REMOVED***;

  // Setup the browsers list
  appPackage.browserslist = defaultBrowsers;

  // Add templatePackage keys/values to appPackage, replacing existing entries
  templatePackageToReplace.forEach(key => {
    appPackage[key] = templatePackage[key];
  ***REMOVED***);

  fs.writeFileSync(
    path.join(appPath, 'package.json'),
    JSON.stringify(appPackage, null, 2) + os.EOL
  );

  const readmeExists = fs.existsSync(path.join(appPath, 'README.md'));
  if (readmeExists) {
    fs.renameSync(
      path.join(appPath, 'README.md'),
      path.join(appPath, 'README.old.md')
    );
  ***REMOVED***

  // Copy the files for the user
  const templateDir = path.join(templatePath, 'template');
  if (fs.existsSync(templateDir)) {
    fs.copySync(templateDir, appPath);
  ***REMOVED*** else {
    console.error(
      `Could not locate supplied template: ${chalk.green(templateDir)***REMOVED***`
    );
    return;
  ***REMOVED***

  // modifies README.md commands based on user used package manager.
  if (useYarn) {
    try {
      const readme = fs.readFileSync(path.join(appPath, 'README.md'), 'utf8');
      fs.writeFileSync(
        path.join(appPath, 'README.md'),
        readme.replace(/(npm run |npm )/g, 'yarn '),
        'utf8'
      );
    ***REMOVED*** catch (err) {
      // Silencing the error. As it fall backs to using default npm commands.
    ***REMOVED***
  ***REMOVED***

  const gitignoreExists = fs.existsSync(path.join(appPath, '.gitignore'));
  if (gitignoreExists) {
    // Append if there's already a `.gitignore` file there
    const data = fs.readFileSync(path.join(appPath, 'gitignore'));
    fs.appendFileSync(path.join(appPath, '.gitignore'), data);
    fs.unlinkSync(path.join(appPath, 'gitignore'));
  ***REMOVED*** else {
    // Rename gitignore after the fact to prevent npm from renaming it to .npmignore
    // See: https://github.com/npm/npm/issues/1862
    fs.moveSync(
      path.join(appPath, 'gitignore'),
      path.join(appPath, '.gitignore'),
      []
    );
  ***REMOVED***

  // Initialize git repo
  let initializedGit = false;

  if (tryGitInit()) {
    initializedGit = true;
    console.log();
    console.log('Initialized a git repository.');
  ***REMOVED***

  let command;
  let remove;
  let args;

  if (useYarn) {
    command = 'yarnpkg';
    remove = 'remove';
    args = ['add'];
  ***REMOVED*** else {
    command = 'npm';
    remove = 'uninstall';
    args = [
      'install',
      '--no-audit', // https://github.com/facebook/create-react-app/issues/11174
      '--save',
      verbose && '--verbose',
    ].filter(e => e);
  ***REMOVED***

  // Install additional template dependencies, if present.
  const dependenciesToInstall = Object.entries({
    ...templatePackage.dependencies,
    ...templatePackage.devDependencies,
  ***REMOVED***);
  if (dependenciesToInstall.length) {
    args = args.concat(
      dependenciesToInstall.map(([dependency, version]) => {
        return `${dependency***REMOVED***@${version***REMOVED***`;
      ***REMOVED***)
    );
  ***REMOVED***

  // Install react and react-dom for backward compatibility with old CRA cli
  // which doesn't install react and react-dom along with react-scripts
  if (!isReactInstalled(appPackage)) {
    args = args.concat(['react', 'react-dom']);
  ***REMOVED***

  // Install template dependencies, and react and react-dom if missing.
  if ((!isReactInstalled(appPackage) || templateName) && args.length > 1) {
    console.log();
    console.log(`Installing template dependencies using ${command***REMOVED***...`);

    const proc = spawn.sync(command, args, { stdio: 'inherit' ***REMOVED***);
    if (proc.status !== 0) {
      console.error(`\`${command***REMOVED*** ${args.join(' ')***REMOVED***\` failed`);
      return;
    ***REMOVED***
  ***REMOVED***

  if (args.find(arg => arg.includes('typescript'))) {
    console.log();
    verifyTypeScriptSetup();
  ***REMOVED***

  // Remove template
  console.log(`Removing template package using ${command***REMOVED***...`);
  console.log();

  const proc = spawn.sync(command, [remove, templateName], {
    stdio: 'inherit',
  ***REMOVED***);
  if (proc.status !== 0) {
    console.error(`\`${command***REMOVED*** ${args.join(' ')***REMOVED***\` failed`);
    return;
  ***REMOVED***

  // Create git commit if git repo was initialized
  if (initializedGit && tryGitCommit(appPath)) {
    console.log();
    console.log('Created git commit.');
  ***REMOVED***

  // Display the most elegant way to cd.
  // This needs to handle an undefined originalDirectory for
  // backward compatibility with old global-cli's.
  let cdpath;
  if (originalDirectory && path.join(originalDirectory, appName) === appPath) {
    cdpath = appName;
  ***REMOVED*** else {
    cdpath = appPath;
  ***REMOVED***

  // Change displayed command to yarn instead of yarnpkg
  const displayedCommand = useYarn ? 'yarn' : 'npm';

  console.log();
  console.log(`Success! Created ${appName***REMOVED*** at ${appPath***REMOVED***`);
  console.log('Inside that directory, you can run several commands:');
  console.log();
  console.log(chalk.cyan(`  ${displayedCommand***REMOVED*** start`));
  console.log('    Starts the development server.');
  console.log();
  console.log(
    chalk.cyan(`  ${displayedCommand***REMOVED*** ${useYarn ? '' : 'run '***REMOVED***build`)
  );
  console.log('    Bundles the app into static files for production.');
  console.log();
  console.log(chalk.cyan(`  ${displayedCommand***REMOVED*** test`));
  console.log('    Starts the test runner.');
  console.log();
  console.log(
    chalk.cyan(`  ${displayedCommand***REMOVED*** ${useYarn ? '' : 'run '***REMOVED***eject`)
  );
  console.log(
    '    Removes this tool and copies build dependencies, configuration files'
  );
  console.log(
    '    and scripts into the app directory. If you do this, you can’t go back!'
  );
  console.log();
  console.log('We suggest that you begin by typing:');
  console.log();
  console.log(chalk.cyan('  cd'), cdpath);
  console.log(`  ${chalk.cyan(`${displayedCommand***REMOVED*** start`)***REMOVED***`);
  if (readmeExists) {
    console.log();
    console.log(
      chalk.yellow(
        'You had a `README.md` file, we renamed it to `README.old.md`'
      )
    );
  ***REMOVED***
  console.log();
  console.log('Happy hacking!');
***REMOVED***;

function isReactInstalled(appPackage) {
  const dependencies = appPackage.dependencies || {***REMOVED***;

  return (
    typeof dependencies.react !== 'undefined' &&
    typeof dependencies['react-dom'] !== 'undefined'
  );
***REMOVED***
