// fork-ts-checker-webpack-plugin/lib/typescript-reporter/file-system/MemFileSystem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const path_1 = require("path");
const memfs_1 = require("memfs");
/**
 * It's an implementation of FileSystem interface which reads and writes to the in-memory file system.
 *
 * @param realFileSystem
 */
function createMemFileSystem(realFileSystem) {
    function exists(path) {
        return memfs_1.fs.existsSync(realFileSystem.normalizePath(path));
    ***REMOVED***
    function readStats(path) {
        return exists(path) ? memfs_1.fs.statSync(realFileSystem.normalizePath(path)) : undefined;
    ***REMOVED***
    function readFile(path, encoding) {
        const stats = readStats(path);
        if (stats && stats.isFile()) {
            return memfs_1.fs
                .readFileSync(realFileSystem.normalizePath(path), { encoding: encoding ***REMOVED***)
                .toString();
        ***REMOVED***
    ***REMOVED***
    function readDir(path) {
        const stats = readStats(path);
        if (stats && stats.isDirectory()) {
            return memfs_1.fs.readdirSync(realFileSystem.normalizePath(path), {
                withFileTypes: true,
            ***REMOVED***);
        ***REMOVED***
        return [];
    ***REMOVED***
    function createDir(path) {
        memfs_1.fs.mkdirSync(realFileSystem.normalizePath(path), { recursive: true ***REMOVED***);
    ***REMOVED***
    function writeFile(path, data) {
        if (!exists(path_1.dirname(path))) {
            createDir(path_1.dirname(path));
        ***REMOVED***
        memfs_1.fs.writeFileSync(realFileSystem.normalizePath(path), data);
    ***REMOVED***
    function deleteFile(path) {
        if (exists(path)) {
            memfs_1.fs.unlinkSync(realFileSystem.normalizePath(path));
        ***REMOVED***
    ***REMOVED***
    function updateTimes(path, atime, mtime) {
        if (exists(path)) {
            memfs_1.fs.utimesSync(realFileSystem.normalizePath(path), atime, mtime);
        ***REMOVED***
    ***REMOVED***
    return Object.assign(Object.assign({***REMOVED***, realFileSystem), { exists(path) {
            return exists(realFileSystem.realPath(path));
        ***REMOVED***,
        readFile(path, encoding) {
            return readFile(realFileSystem.realPath(path), encoding);
        ***REMOVED***,
        readDir(path) {
            return readDir(realFileSystem.realPath(path));
        ***REMOVED***,
        readStats(path) {
            return readStats(realFileSystem.realPath(path));
        ***REMOVED***,
        writeFile(path, data) {
            writeFile(realFileSystem.realPath(path), data);
        ***REMOVED***,
        deleteFile(path) {
            deleteFile(realFileSystem.realPath(path));
        ***REMOVED***,
        createDir(path) {
            createDir(realFileSystem.realPath(path));
        ***REMOVED***,
        updateTimes(path, atime, mtime) {
            updateTimes(realFileSystem.realPath(path), atime, mtime);
        ***REMOVED***,
        clearCache() {
            realFileSystem.clearCache();
        ***REMOVED*** ***REMOVED***);
***REMOVED***
exports.createMemFileSystem = createMemFileSystem;
