// fork-ts-checker-webpack-plugin/lib/typescript-reporter/file-system/RealFileSystem.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod ***REMOVED***;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const path_1 = require("path");
const fs_extra_1 = __importDefault(require("fs-extra"));
/**
 * It's an implementation of the FileSystem interface which reads and writes directly to the real file system.
 *
 * @param caseSensitive
 */
function createRealFileSystem(caseSensitive = false) {
    // read cache
    const existsCache = new Map();
    const readStatsCache = new Map();
    const readFileCache = new Map();
    const readDirCache = new Map();
    const realPathCache = new Map();
    function normalizePath(path) {
        return caseSensitive ? path_1.normalize(path) : path_1.normalize(path).toLowerCase();
    ***REMOVED***
    // read methods
    function exists(path) {
        const normalizedPath = normalizePath(path);
        if (!existsCache.has(normalizedPath)) {
            existsCache.set(normalizedPath, fs_extra_1.default.existsSync(normalizedPath));
        ***REMOVED***
        return !!existsCache.get(normalizedPath);
    ***REMOVED***
    function readStats(path) {
        const normalizedPath = normalizePath(path);
        if (!readStatsCache.has(normalizedPath)) {
            if (exists(normalizedPath)) {
                readStatsCache.set(normalizedPath, fs_extra_1.default.statSync(normalizedPath));
            ***REMOVED***
        ***REMOVED***
        return readStatsCache.get(normalizedPath);
    ***REMOVED***
    function readFile(path, encoding) {
        const normalizedPath = normalizePath(path);
        if (!readFileCache.has(normalizedPath)) {
            const stats = readStats(normalizedPath);
            if (stats && stats.isFile()) {
                readFileCache.set(normalizedPath, fs_extra_1.default.readFileSync(normalizedPath, { encoding: encoding ***REMOVED***).toString());
            ***REMOVED***
            else {
                readFileCache.set(normalizedPath, undefined);
            ***REMOVED***
        ***REMOVED***
        return readFileCache.get(normalizedPath);
    ***REMOVED***
    function readDir(path) {
        const normalizedPath = normalizePath(path);
        if (!readDirCache.has(normalizedPath)) {
            const stats = readStats(normalizedPath);
            if (stats && stats.isDirectory()) {
                readDirCache.set(normalizedPath, fs_extra_1.default.readdirSync(normalizedPath, { withFileTypes: true ***REMOVED***));
            ***REMOVED***
            else {
                readDirCache.set(normalizedPath, []);
            ***REMOVED***
        ***REMOVED***
        return readDirCache.get(normalizedPath) || [];
    ***REMOVED***
    function getRealPath(path) {
        const normalizedPath = normalizePath(path);
        if (!realPathCache.has(normalizedPath)) {
            let base = normalizedPath;
            let nested = '';
            while (base !== path_1.dirname(base)) {
                if (exists(base)) {
                    realPathCache.set(normalizedPath, normalizePath(path_1.join(fs_extra_1.default.realpathSync(base), nested)));
                    break;
                ***REMOVED***
                nested = path_1.join(path_1.basename(base), nested);
                base = path_1.dirname(base);
            ***REMOVED***
        ***REMOVED***
        return realPathCache.get(normalizedPath) || normalizedPath;
    ***REMOVED***
    function createDir(path) {
        const normalizedPath = normalizePath(path);
        fs_extra_1.default.mkdirSync(normalizedPath, { recursive: true ***REMOVED***);
        // update cache
        existsCache.set(normalizedPath, true);
        if (readDirCache.has(path_1.dirname(normalizedPath))) {
            readDirCache.delete(path_1.dirname(normalizedPath));
        ***REMOVED***
        if (readStatsCache.has(normalizedPath)) {
            readStatsCache.delete(normalizedPath);
        ***REMOVED***
    ***REMOVED***
    function writeFile(path, data) {
        const normalizedPath = normalizePath(path);
        if (!exists(path_1.dirname(normalizedPath))) {
            createDir(path_1.dirname(normalizedPath));
        ***REMOVED***
        fs_extra_1.default.writeFileSync(normalizedPath, data);
        // update cache
        existsCache.set(normalizedPath, true);
        if (readDirCache.has(path_1.dirname(normalizedPath))) {
            readDirCache.delete(path_1.dirname(normalizedPath));
        ***REMOVED***
        if (readStatsCache.has(normalizedPath)) {
            readStatsCache.delete(normalizedPath);
        ***REMOVED***
        if (readFileCache.has(normalizedPath)) {
            readFileCache.delete(normalizedPath);
        ***REMOVED***
    ***REMOVED***
    function deleteFile(path) {
        if (exists(path)) {
            const normalizedPath = normalizePath(path);
            fs_extra_1.default.unlinkSync(normalizedPath);
            // update cache
            existsCache.set(normalizedPath, false);
            if (readDirCache.has(path_1.dirname(normalizedPath))) {
                readDirCache.delete(path_1.dirname(normalizedPath));
            ***REMOVED***
            if (readStatsCache.has(normalizedPath)) {
                readStatsCache.delete(normalizedPath);
            ***REMOVED***
            if (readFileCache.has(normalizedPath)) {
                readFileCache.delete(normalizedPath);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    function updateTimes(path, atime, mtime) {
        if (exists(path)) {
            const normalizedPath = normalizePath(path);
            fs_extra_1.default.utimesSync(normalizePath(path), atime, mtime);
            // update cache
            if (readStatsCache.has(normalizedPath)) {
                readStatsCache.delete(normalizedPath);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    return {
        exists(path) {
            return exists(getRealPath(path));
        ***REMOVED***,
        readFile(path, encoding) {
            return readFile(getRealPath(path), encoding);
        ***REMOVED***,
        readDir(path) {
            return readDir(getRealPath(path));
        ***REMOVED***,
        readStats(path) {
            return readStats(getRealPath(path));
        ***REMOVED***,
        realPath(path) {
            return getRealPath(path);
        ***REMOVED***,
        normalizePath(path) {
            return normalizePath(path);
        ***REMOVED***,
        writeFile(path, data) {
            writeFile(getRealPath(path), data);
        ***REMOVED***,
        deleteFile(path) {
            deleteFile(getRealPath(path));
        ***REMOVED***,
        createDir(path) {
            createDir(getRealPath(path));
        ***REMOVED***,
        updateTimes(path, atime, mtime) {
            updateTimes(getRealPath(path), atime, mtime);
        ***REMOVED***,
        clearCache() {
            existsCache.clear();
            readStatsCache.clear();
            readFileCache.clear();
            readDirCache.clear();
            realPathCache.clear();
        ***REMOVED***,
    ***REMOVED***;
***REMOVED***
exports.createRealFileSystem = createRealFileSystem;
