// jsdom/lib/jsdom/living/range/AbstractRange-impl.js
"use strict";

// https://dom.spec.whatwg.org/#abstractrange
class AbstractRangeImpl {
  constructor(globalObject, args, privateData) {
    const { start, end ***REMOVED*** = privateData;

    this._start = start;
    this._end = end;

    this._globalObject = globalObject;
  ***REMOVED***

  // https://dom.spec.whatwg.org/#dom-range-startcontainer
  get startContainer() {
    return this._start.node;
  ***REMOVED***

  // https://dom.spec.whatwg.org/#dom-range-startoffset
  get startOffset() {
    return this._start.offset;
  ***REMOVED***

  // https://dom.spec.whatwg.org/#dom-range-endcontainer
  get endContainer() {
    return this._end.node;
  ***REMOVED***

  // https://dom.spec.whatwg.org/#dom-range-endoffset
  get endOffset() {
    return this._end.offset;
  ***REMOVED***

  // https://dom.spec.whatwg.org/#dom-range-collapsed
  get collapsed() {
    const { _start, _end ***REMOVED*** = this;
    return _start.node === _end.node && _start.offset === _end.offset;
  ***REMOVED***
***REMOVED***

module.exports = {
  implementation: AbstractRangeImpl
***REMOVED***;
