// fork-ts-checker-webpack-plugin/lib/typescript-reporter/reporter/ControlledTypeScriptSystem.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); ***REMOVED***); ***REMOVED***
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); ***REMOVED*** catch (e) { reject(e); ***REMOVED*** ***REMOVED***
        function rejected(value) { try { step(generator["throw"](value)); ***REMOVED*** catch (e) { reject(e); ***REMOVED*** ***REMOVED***
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); ***REMOVED***
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    ***REMOVED***);
***REMOVED***;
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod ***REMOVED***;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const path_1 = require("path");
const PassiveFileSystem_1 = require("../file-system/PassiveFileSystem");
const forwardSlash_1 = __importDefault(require("../../utils/path/forwardSlash"));
const RealFileSystem_1 = require("../file-system/RealFileSystem");
const MemFileSystem_1 = require("../file-system/MemFileSystem");
function createControlledTypeScriptSystem(typescript, mode = 'readonly') {
    let artifacts = {
        files: [],
        dirs: [],
        excluded: [],
        extensions: [],
    ***REMOVED***;
    let isInitialRun = true;
    // watchers
    const fileWatcherCallbacksMap = new Map();
    const directoryWatcherCallbacksMap = new Map();
    const recursiveDirectoryWatcherCallbacksMap = new Map();
    const deletedFiles = new Map();
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const timeoutCallbacks = new Set();
    // always use case-sensitive as normalization to lower-case can be a problem for some
    // third-party libraries, like fsevents
    const caseSensitive = true;
    const realFileSystem = RealFileSystem_1.createRealFileSystem(caseSensitive);
    const memFileSystem = MemFileSystem_1.createMemFileSystem(realFileSystem);
    const passiveFileSystem = PassiveFileSystem_1.createPassiveFileSystem(memFileSystem, realFileSystem);
    // based on the ts.ignorePaths
    const ignoredPaths = ['/node_modules/.', '/.git', '/.#'];
    function createWatcher(watchersMap, path, callback) {
        const normalizedPath = realFileSystem.normalizePath(path);
        const watchers = watchersMap.get(normalizedPath) || [];
        const nextWatchers = [...watchers, callback];
        watchersMap.set(normalizedPath, nextWatchers);
        return {
            close: () => {
                const watchers = watchersMap.get(normalizedPath) || [];
                const nextWatchers = watchers.filter((watcher) => watcher !== callback);
                if (nextWatchers.length > 0) {
                    watchersMap.set(normalizedPath, nextWatchers);
                ***REMOVED***
                else {
                    watchersMap.delete(normalizedPath);
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***;
    ***REMOVED***
    function invokeFileWatchers(path, event) {
        const normalizedPath = realFileSystem.normalizePath(path);
        if (normalizedPath.endsWith('.js')) {
            // trigger relevant .d.ts file watcher - handles the case, when we have webpack watcher
            // that points to a symlinked package
            invokeFileWatchers(normalizedPath.slice(0, -3) + '.d.ts', event);
        ***REMOVED***
        const fileWatcherCallbacks = fileWatcherCallbacksMap.get(normalizedPath);
        if (fileWatcherCallbacks) {
            // typescript expects normalized paths with posix forward slash
            fileWatcherCallbacks.forEach((fileWatcherCallback) => fileWatcherCallback(forwardSlash_1.default(normalizedPath), event));
        ***REMOVED***
    ***REMOVED***
    function invokeDirectoryWatchers(path) {
        const normalizedPath = realFileSystem.normalizePath(path);
        const directory = path_1.dirname(normalizedPath);
        if (ignoredPaths.some((ignoredPath) => forwardSlash_1.default(normalizedPath).includes(ignoredPath))) {
            return;
        ***REMOVED***
        const directoryWatcherCallbacks = directoryWatcherCallbacksMap.get(directory);
        if (directoryWatcherCallbacks) {
            directoryWatcherCallbacks.forEach((directoryWatcherCallback) => directoryWatcherCallback(forwardSlash_1.default(normalizedPath)));
        ***REMOVED***
        recursiveDirectoryWatcherCallbacksMap.forEach((recursiveDirectoryWatcherCallbacks, watchedDirectory) => {
            if (watchedDirectory === directory ||
                (directory.startsWith(watchedDirectory) &&
                    forwardSlash_1.default(directory)[watchedDirectory.length] === '/')) {
                recursiveDirectoryWatcherCallbacks.forEach((recursiveDirectoryWatcherCallback) => recursiveDirectoryWatcherCallback(forwardSlash_1.default(normalizedPath)));
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***
    function isArtifact(path) {
        return ((artifacts.dirs.some((dir) => path.includes(dir)) ||
            artifacts.files.some((file) => path === file)) &&
            artifacts.extensions.some((extension) => path.endsWith(extension)));
    ***REMOVED***
    function getReadFileSystem(path) {
        if ((mode === 'readonly' || mode === 'write-tsbuildinfo') && isArtifact(path)) {
            if (isInitialRun && !memFileSystem.exists(path) && passiveFileSystem.exists(path)) {
                // copy file to memory on initial run
                const stats = passiveFileSystem.readStats(path);
                if (stats === null || stats === void 0 ? void 0 : stats.isFile()) {
                    const content = passiveFileSystem.readFile(path);
                    if (content) {
                        memFileSystem.writeFile(path, content);
                        memFileSystem.updateTimes(path, stats.atime, stats.mtime);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
            return memFileSystem;
        ***REMOVED***
        return passiveFileSystem;
    ***REMOVED***
    function getWriteFileSystem(path) {
        if (mode === 'write-references' ||
            (mode === 'write-tsbuildinfo' && path.endsWith('.tsbuildinfo')) ||
            (mode === 'write-dts' &&
                ['.tsbuildinfo', '.d.ts', '.d.ts.map'].some((prefix) => path.endsWith(prefix)))) {
            return realFileSystem;
        ***REMOVED***
        return passiveFileSystem;
    ***REMOVED***
    const controlledSystem = Object.assign(Object.assign({***REMOVED***, typescript.sys), { useCaseSensitiveFileNames: caseSensitive, realpath(path) {
            return getReadFileSystem(path).realPath(path);
        ***REMOVED***,
        fileExists(path) {
            const stats = getReadFileSystem(path).readStats(path);
            return !!stats && stats.isFile();
        ***REMOVED***,
        readFile(path, encoding) {
            return getReadFileSystem(path).readFile(path, encoding);
        ***REMOVED***,
        getFileSize(path) {
            const stats = getReadFileSystem(path).readStats(path);
            return stats ? stats.size : 0;
        ***REMOVED***,
        writeFile(path, data) {
            getWriteFileSystem(path).writeFile(path, data);
            controlledSystem.invokeFileChanged(path);
        ***REMOVED***,
        deleteFile(path) {
            getWriteFileSystem(path).deleteFile(path);
            controlledSystem.invokeFileDeleted(path);
        ***REMOVED***,
        directoryExists(path) {
            var _a;
            return Boolean((_a = getReadFileSystem(path).readStats(path)) === null || _a === void 0 ? void 0 : _a.isDirectory());
        ***REMOVED***,
        createDirectory(path) {
            getWriteFileSystem(path).createDir(path);
            invokeDirectoryWatchers(path);
        ***REMOVED***,
        getDirectories(path) {
            const dirents = getReadFileSystem(path).readDir(path);
            return dirents
                .filter((dirent) => dirent.isDirectory() ||
                (dirent.isSymbolicLink() && controlledSystem.directoryExists(path_1.join(path, dirent.name))))
                .map((dirent) => dirent.name);
        ***REMOVED***,
        getModifiedTime(path) {
            const stats = getReadFileSystem(path).readStats(path);
            if (stats) {
                return stats.mtime;
            ***REMOVED***
        ***REMOVED***,
        setModifiedTime(path, date) {
            getWriteFileSystem(path).updateTimes(path, date, date);
            invokeDirectoryWatchers(path);
            invokeFileWatchers(path, typescript.FileWatcherEventKind.Changed);
        ***REMOVED***,
        watchFile(path, callback) {
            return createWatcher(fileWatcherCallbacksMap, path, callback);
        ***REMOVED***,
        watchDirectory(path, callback, recursive = false) {
            return createWatcher(recursive ? recursiveDirectoryWatcherCallbacksMap : directoryWatcherCallbacksMap, path, callback);
        ***REMOVED***, 
        // use immediate instead of timeout to avoid waiting 250ms for batching files changes
        setTimeout: (callback, timeout, ...args) => {
            const timeoutId = setImmediate(() => {
                callback(...args);
                timeoutCallbacks.delete(timeoutId);
            ***REMOVED***);
            timeoutCallbacks.add(timeoutId);
            return timeoutId;
        ***REMOVED***, clearTimeout: (timeoutId) => {
            clearImmediate(timeoutId);
            timeoutCallbacks.delete(timeoutId);
        ***REMOVED***, waitForQueued() {
            return __awaiter(this, void 0, void 0, function* () {
                while (timeoutCallbacks.size > 0) {
                    yield new Promise((resolve) => setImmediate(resolve));
                ***REMOVED***
                isInitialRun = false;
            ***REMOVED***);
        ***REMOVED***,
        invokeFileCreated(path) {
            const normalizedPath = realFileSystem.normalizePath(path);
            invokeFileWatchers(path, typescript.FileWatcherEventKind.Created);
            invokeDirectoryWatchers(normalizedPath);
            deletedFiles.set(normalizedPath, false);
        ***REMOVED***,
        invokeFileChanged(path) {
            const normalizedPath = realFileSystem.normalizePath(path);
            if (deletedFiles.get(normalizedPath) || !fileWatcherCallbacksMap.has(normalizedPath)) {
                invokeFileWatchers(path, typescript.FileWatcherEventKind.Created);
                invokeDirectoryWatchers(normalizedPath);
                deletedFiles.set(normalizedPath, false);
            ***REMOVED***
            else {
                invokeFileWatchers(path, typescript.FileWatcherEventKind.Changed);
            ***REMOVED***
        ***REMOVED***,
        invokeFileDeleted(path) {
            const normalizedPath = realFileSystem.normalizePath(path);
            if (!deletedFiles.get(normalizedPath)) {
                invokeFileWatchers(path, typescript.FileWatcherEventKind.Deleted);
                invokeDirectoryWatchers(path);
                deletedFiles.set(normalizedPath, true);
            ***REMOVED***
        ***REMOVED***,
        clearCache() {
            realFileSystem.clearCache();
            memFileSystem.clearCache();
            passiveFileSystem.clearCache();
        ***REMOVED***,
        setArtifacts(nextArtifacts) {
            artifacts = nextArtifacts;
        ***REMOVED*** ***REMOVED***);
    return controlledSystem;
***REMOVED***
exports.createControlledTypeScriptSystem = createControlledTypeScriptSystem;
