// fork-ts-checker-webpack-plugin/lib/typescript-reporter/reporter/ControlledWatchSolutionBuilderHost.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const ControlledWatchCompilerHost_1 = require("./ControlledWatchCompilerHost");
function createControlledWatchSolutionBuilderHost(typescript, parsedCommandLine, system, createProgram, reportDiagnostic, reportWatchStatus, reportSolutionBuilderStatus, afterProgramCreate, afterProgramEmitAndDiagnostics, hostExtensions = []) {
    const controlledWatchCompilerHost = ControlledWatchCompilerHost_1.createControlledWatchCompilerHost(typescript, parsedCommandLine, system, createProgram, reportDiagnostic, reportWatchStatus, afterProgramCreate, hostExtensions);
    let controlledWatchSolutionBuilderHost = Object.assign(Object.assign({***REMOVED***, controlledWatchCompilerHost), { reportDiagnostic(diagnostic) {
            if (reportDiagnostic) {
                reportDiagnostic(diagnostic);
            ***REMOVED***
        ***REMOVED***,
        reportSolutionBuilderStatus(diagnostic) {
            if (reportSolutionBuilderStatus) {
                reportSolutionBuilderStatus(diagnostic);
            ***REMOVED***
        ***REMOVED***,
        afterProgramEmitAndDiagnostics(program) {
            if (afterProgramEmitAndDiagnostics) {
                afterProgramEmitAndDiagnostics(program);
            ***REMOVED***
        ***REMOVED***,
        createDirectory(path) {
            system.createDirectory(path);
        ***REMOVED***,
        writeFile(path, data) {
            system.writeFile(path, data);
        ***REMOVED***,
        getModifiedTime(fileName) {
            return system.getModifiedTime(fileName);
        ***REMOVED***,
        setModifiedTime(fileName, date) {
            system.setModifiedTime(fileName, date);
        ***REMOVED***,
        deleteFile(fileName) {
            system.deleteFile(fileName);
        ***REMOVED***,
        getParsedCommandLine(fileName) {
            return typescript.getParsedCommandLineOfConfigFile(fileName, { skipLibCheck: true ***REMOVED***, Object.assign(Object.assign({***REMOVED***, system), { onUnRecoverableConfigFileDiagnostic: (diagnostic) => {
                    if (reportDiagnostic) {
                        reportDiagnostic(diagnostic);
                    ***REMOVED***
                ***REMOVED*** ***REMOVED***));
        ***REMOVED*** ***REMOVED***);
    hostExtensions.forEach((hostExtension) => {
        if (hostExtension.extendWatchSolutionBuilderHost) {
            controlledWatchSolutionBuilderHost = hostExtension.extendWatchSolutionBuilderHost(controlledWatchSolutionBuilderHost, parsedCommandLine);
        ***REMOVED***
    ***REMOVED***);
    return controlledWatchSolutionBuilderHost;
***REMOVED***
exports.createControlledWatchSolutionBuilderHost = createControlledWatchSolutionBuilderHost;
