// sucrase/dist/esm/parser/plugins/flow.js
/* eslint max-len: 0 */

import {
  eat,
  lookaheadType,
  lookaheadTypeAndKeyword,
  match,
  next,
  popTypeContext,
  pushTypeContext,

***REMOVED*** from "../tokenizer/index";
import {ContextualKeyword***REMOVED*** from "../tokenizer/keywords";
import {TokenType, TokenType as tt***REMOVED*** from "../tokenizer/types";
import {input, state***REMOVED*** from "../traverser/base";
import {
  baseParseMaybeAssign,
  baseParseSubscript,
  baseParseSubscripts,
  parseArrow,
  parseArrowExpression,
  parseCallExpressionArguments,
  parseExprAtom,
  parseExpression,
  parseFunctionBody,
  parseIdentifier,
  parseLiteral,

***REMOVED*** from "../traverser/expression";
import {
  baseParseExportStar,
  parseExport,
  parseExportFrom,
  parseExportSpecifiers,
  parseFunctionParams,
  parseImport,
  parseStatement,
***REMOVED*** from "../traverser/statement";
import {
  canInsertSemicolon,
  eatContextual,
  expect,
  expectContextual,
  isContextual,
  isLookaheadContextual,
  semicolon,
  unexpected,
***REMOVED*** from "../traverser/util";

function isMaybeDefaultImport(lookahead) {
  return (
    (lookahead.type === tt.name || !!(lookahead.type & TokenType.IS_KEYWORD)) &&
    lookahead.contextualKeyword !== ContextualKeyword._from
  );
***REMOVED***

function flowParseTypeInitialiser(tok) {
  const oldIsType = pushTypeContext(0);
  expect(tok || tt.colon);
  flowParseType();
  popTypeContext(oldIsType);
***REMOVED***

function flowParsePredicate() {
  expect(tt.modulo);
  expectContextual(ContextualKeyword._checks);
  if (eat(tt.parenL)) {
    parseExpression();
    expect(tt.parenR);
  ***REMOVED***
***REMOVED***

function flowParseTypeAndPredicateInitialiser() {
  const oldIsType = pushTypeContext(0);
  expect(tt.colon);
  if (match(tt.modulo)) {
    flowParsePredicate();
  ***REMOVED*** else {
    flowParseType();
    if (match(tt.modulo)) {
      flowParsePredicate();
    ***REMOVED***
  ***REMOVED***
  popTypeContext(oldIsType);
***REMOVED***

function flowParseDeclareClass() {
  next();
  flowParseInterfaceish(/* isClass */ true);
***REMOVED***

function flowParseDeclareFunction() {
  next();
  parseIdentifier();

  if (match(tt.lessThan)) {
    flowParseTypeParameterDeclaration();
  ***REMOVED***

  expect(tt.parenL);
  flowParseFunctionTypeParams();
  expect(tt.parenR);

  flowParseTypeAndPredicateInitialiser();

  semicolon();
***REMOVED***

function flowParseDeclare() {
  if (match(tt._class)) {
    flowParseDeclareClass();
  ***REMOVED*** else if (match(tt._function)) {
    flowParseDeclareFunction();
  ***REMOVED*** else if (match(tt._var)) {
    flowParseDeclareVariable();
  ***REMOVED*** else if (eatContextual(ContextualKeyword._module)) {
    if (eat(tt.dot)) {
      flowParseDeclareModuleExports();
    ***REMOVED*** else {
      flowParseDeclareModule();
    ***REMOVED***
  ***REMOVED*** else if (isContextual(ContextualKeyword._type)) {
    flowParseDeclareTypeAlias();
  ***REMOVED*** else if (isContextual(ContextualKeyword._opaque)) {
    flowParseDeclareOpaqueType();
  ***REMOVED*** else if (isContextual(ContextualKeyword._interface)) {
    flowParseDeclareInterface();
  ***REMOVED*** else if (match(tt._export)) {
    flowParseDeclareExportDeclaration();
  ***REMOVED*** else {
    unexpected();
  ***REMOVED***
***REMOVED***

function flowParseDeclareVariable() {
  next();
  flowParseTypeAnnotatableIdentifier();
  semicolon();
***REMOVED***

function flowParseDeclareModule() {
  if (match(tt.string)) {
    parseExprAtom();
  ***REMOVED*** else {
    parseIdentifier();
  ***REMOVED***

  expect(tt.braceL);
  while (!match(tt.braceR) && !state.error) {
    if (match(tt._import)) {
      next();
      parseImport();
    ***REMOVED*** else {
      unexpected();
    ***REMOVED***
  ***REMOVED***
  expect(tt.braceR);
***REMOVED***

function flowParseDeclareExportDeclaration() {
  expect(tt._export);

  if (eat(tt._default)) {
    if (match(tt._function) || match(tt._class)) {
      // declare export default class ...
      // declare export default function ...
      flowParseDeclare();
    ***REMOVED*** else {
      // declare export default [type];
      flowParseType();
      semicolon();
    ***REMOVED***
  ***REMOVED*** else if (
    match(tt._var) || // declare export var ...
    match(tt._function) || // declare export function ...
    match(tt._class) || // declare export class ...
    isContextual(ContextualKeyword._opaque) // declare export opaque ..
  ) {
    flowParseDeclare();
  ***REMOVED*** else if (
    match(tt.star) || // declare export * from ''
    match(tt.braceL) || // declare export {***REMOVED*** ...
    isContextual(ContextualKeyword._interface) || // declare export interface ...
    isContextual(ContextualKeyword._type) || // declare export type ...
    isContextual(ContextualKeyword._opaque) // declare export opaque type ...
  ) {
    parseExport();
  ***REMOVED*** else {
    unexpected();
  ***REMOVED***
***REMOVED***

function flowParseDeclareModuleExports() {
  expectContextual(ContextualKeyword._exports);
  flowParseTypeAnnotation();
  semicolon();
***REMOVED***

function flowParseDeclareTypeAlias() {
  next();
  flowParseTypeAlias();
***REMOVED***

function flowParseDeclareOpaqueType() {
  next();
  flowParseOpaqueType(true);
***REMOVED***

function flowParseDeclareInterface() {
  next();
  flowParseInterfaceish();
***REMOVED***

// Interfaces

function flowParseInterfaceish(isClass = false) {
  flowParseRestrictedIdentifier();

  if (match(tt.lessThan)) {
    flowParseTypeParameterDeclaration();
  ***REMOVED***

  if (eat(tt._extends)) {
    do {
      flowParseInterfaceExtends();
    ***REMOVED*** while (!isClass && eat(tt.comma));
  ***REMOVED***

  if (isContextual(ContextualKeyword._mixins)) {
    next();
    do {
      flowParseInterfaceExtends();
    ***REMOVED*** while (eat(tt.comma));
  ***REMOVED***

  if (isContextual(ContextualKeyword._implements)) {
    next();
    do {
      flowParseInterfaceExtends();
    ***REMOVED*** while (eat(tt.comma));
  ***REMOVED***

  flowParseObjectType(isClass, false, isClass);
***REMOVED***

function flowParseInterfaceExtends() {
  flowParseQualifiedTypeIdentifier(false);
  if (match(tt.lessThan)) {
    flowParseTypeParameterInstantiation();
  ***REMOVED***
***REMOVED***

function flowParseInterface() {
  flowParseInterfaceish();
***REMOVED***

function flowParseRestrictedIdentifier() {
  parseIdentifier();
***REMOVED***

function flowParseTypeAlias() {
  flowParseRestrictedIdentifier();

  if (match(tt.lessThan)) {
    flowParseTypeParameterDeclaration();
  ***REMOVED***

  flowParseTypeInitialiser(tt.eq);
  semicolon();
***REMOVED***

function flowParseOpaqueType(declare) {
  expectContextual(ContextualKeyword._type);
  flowParseRestrictedIdentifier();

  if (match(tt.lessThan)) {
    flowParseTypeParameterDeclaration();
  ***REMOVED***

  // Parse the supertype
  if (match(tt.colon)) {
    flowParseTypeInitialiser(tt.colon);
  ***REMOVED***

  if (!declare) {
    flowParseTypeInitialiser(tt.eq);
  ***REMOVED***
  semicolon();
***REMOVED***

function flowParseTypeParameter() {
  flowParseVariance();
  flowParseTypeAnnotatableIdentifier();

  if (eat(tt.eq)) {
    flowParseType();
  ***REMOVED***
***REMOVED***

export function flowParseTypeParameterDeclaration() {
  const oldIsType = pushTypeContext(0);
  // istanbul ignore else: this condition is already checked at all call sites
  if (match(tt.lessThan) || match(tt.typeParameterStart)) {
    next();
  ***REMOVED*** else {
    unexpected();
  ***REMOVED***

  do {
    flowParseTypeParameter();
    if (!match(tt.greaterThan)) {
      expect(tt.comma);
    ***REMOVED***
  ***REMOVED*** while (!match(tt.greaterThan) && !state.error);
  expect(tt.greaterThan);
  popTypeContext(oldIsType);
***REMOVED***

function flowParseTypeParameterInstantiation() {
  const oldIsType = pushTypeContext(0);
  expect(tt.lessThan);
  while (!match(tt.greaterThan) && !state.error) {
    flowParseType();
    if (!match(tt.greaterThan)) {
      expect(tt.comma);
    ***REMOVED***
  ***REMOVED***
  expect(tt.greaterThan);
  popTypeContext(oldIsType);
***REMOVED***

function flowParseInterfaceType() {
  expectContextual(ContextualKeyword._interface);
  if (eat(tt._extends)) {
    do {
      flowParseInterfaceExtends();
    ***REMOVED*** while (eat(tt.comma));
  ***REMOVED***
  flowParseObjectType(false, false, false);
***REMOVED***

function flowParseObjectPropertyKey() {
  if (match(tt.num) || match(tt.string)) {
    parseExprAtom();
  ***REMOVED*** else {
    parseIdentifier();
  ***REMOVED***
***REMOVED***

function flowParseObjectTypeIndexer() {
  // Note: bracketL has already been consumed
  if (lookaheadType() === tt.colon) {
    flowParseObjectPropertyKey();
    flowParseTypeInitialiser();
  ***REMOVED*** else {
    flowParseType();
  ***REMOVED***
  expect(tt.bracketR);
  flowParseTypeInitialiser();
***REMOVED***

function flowParseObjectTypeInternalSlot() {
  // Note: both bracketL have already been consumed
  flowParseObjectPropertyKey();
  expect(tt.bracketR);
  expect(tt.bracketR);
  if (match(tt.lessThan) || match(tt.parenL)) {
    flowParseObjectTypeMethodish();
  ***REMOVED*** else {
    eat(tt.question);
    flowParseTypeInitialiser();
  ***REMOVED***
***REMOVED***

function flowParseObjectTypeMethodish() {
  if (match(tt.lessThan)) {
    flowParseTypeParameterDeclaration();
  ***REMOVED***

  expect(tt.parenL);
  while (!match(tt.parenR) && !match(tt.ellipsis) && !state.error) {
    flowParseFunctionTypeParam();
    if (!match(tt.parenR)) {
      expect(tt.comma);
    ***REMOVED***
  ***REMOVED***

  if (eat(tt.ellipsis)) {
    flowParseFunctionTypeParam();
  ***REMOVED***
  expect(tt.parenR);
  flowParseTypeInitialiser();
***REMOVED***

function flowParseObjectTypeCallProperty() {
  flowParseObjectTypeMethodish();
***REMOVED***

function flowParseObjectType(allowStatic, allowExact, allowProto) {
  let endDelim;
  if (allowExact && match(tt.braceBarL)) {
    expect(tt.braceBarL);
    endDelim = tt.braceBarR;
  ***REMOVED*** else {
    expect(tt.braceL);
    endDelim = tt.braceR;
  ***REMOVED***

  while (!match(endDelim) && !state.error) {
    if (allowProto && isContextual(ContextualKeyword._proto)) {
      const lookahead = lookaheadType();
      if (lookahead !== tt.colon && lookahead !== tt.question) {
        next();
        allowStatic = false;
      ***REMOVED***
    ***REMOVED***
    if (allowStatic && isContextual(ContextualKeyword._static)) {
      const lookahead = lookaheadType();
      if (lookahead !== tt.colon && lookahead !== tt.question) {
        next();
      ***REMOVED***
    ***REMOVED***

    flowParseVariance();

    if (eat(tt.bracketL)) {
      if (eat(tt.bracketL)) {
        flowParseObjectTypeInternalSlot();
      ***REMOVED*** else {
        flowParseObjectTypeIndexer();
      ***REMOVED***
    ***REMOVED*** else if (match(tt.parenL) || match(tt.lessThan)) {
      flowParseObjectTypeCallProperty();
    ***REMOVED*** else {
      if (isContextual(ContextualKeyword._get) || isContextual(ContextualKeyword._set)) {
        const lookahead = lookaheadType();
        if (lookahead === tt.name || lookahead === tt.string || lookahead === tt.num) {
          next();
        ***REMOVED***
      ***REMOVED***

      flowParseObjectTypeProperty();
    ***REMOVED***

    flowObjectTypeSemicolon();
  ***REMOVED***

  expect(endDelim);
***REMOVED***

function flowParseObjectTypeProperty() {
  if (match(tt.ellipsis)) {
    expect(tt.ellipsis);
    if (!eat(tt.comma)) {
      eat(tt.semi);
    ***REMOVED***
    // Explicit inexact object syntax.
    if (match(tt.braceR)) {
      return;
    ***REMOVED***
    flowParseType();
  ***REMOVED*** else {
    flowParseObjectPropertyKey();
    if (match(tt.lessThan) || match(tt.parenL)) {
      // This is a method property
      flowParseObjectTypeMethodish();
    ***REMOVED*** else {
      eat(tt.question);
      flowParseTypeInitialiser();
    ***REMOVED***
  ***REMOVED***
***REMOVED***

function flowObjectTypeSemicolon() {
  if (!eat(tt.semi) && !eat(tt.comma) && !match(tt.braceR) && !match(tt.braceBarR)) {
    unexpected();
  ***REMOVED***
***REMOVED***

function flowParseQualifiedTypeIdentifier(initialIdAlreadyParsed) {
  if (!initialIdAlreadyParsed) {
    parseIdentifier();
  ***REMOVED***
  while (eat(tt.dot)) {
    parseIdentifier();
  ***REMOVED***
***REMOVED***

function flowParseGenericType() {
  flowParseQualifiedTypeIdentifier(true);
  if (match(tt.lessThan)) {
    flowParseTypeParameterInstantiation();
  ***REMOVED***
***REMOVED***

function flowParseTypeofType() {
  expect(tt._typeof);
  flowParsePrimaryType();
***REMOVED***

function flowParseTupleType() {
  expect(tt.bracketL);
  // We allow trailing commas
  while (state.pos < input.length && !match(tt.bracketR)) {
    flowParseType();
    if (match(tt.bracketR)) {
      break;
    ***REMOVED***
    expect(tt.comma);
  ***REMOVED***
  expect(tt.bracketR);
***REMOVED***

function flowParseFunctionTypeParam() {
  const lookahead = lookaheadType();
  if (lookahead === tt.colon || lookahead === tt.question) {
    parseIdentifier();
    eat(tt.question);
    flowParseTypeInitialiser();
  ***REMOVED*** else {
    flowParseType();
  ***REMOVED***
***REMOVED***

function flowParseFunctionTypeParams() {
  while (!match(tt.parenR) && !match(tt.ellipsis) && !state.error) {
    flowParseFunctionTypeParam();
    if (!match(tt.parenR)) {
      expect(tt.comma);
    ***REMOVED***
  ***REMOVED***
  if (eat(tt.ellipsis)) {
    flowParseFunctionTypeParam();
  ***REMOVED***
***REMOVED***

// The parsing of types roughly parallels the parsing of expressions, and
// primary types are kind of like primary expressions...they're the
// primitives with which other types are constructed.
function flowParsePrimaryType() {
  let isGroupedType = false;
  const oldNoAnonFunctionType = state.noAnonFunctionType;

  switch (state.type) {
    case tt.name: {
      if (isContextual(ContextualKeyword._interface)) {
        flowParseInterfaceType();
        return;
      ***REMOVED***
      parseIdentifier();
      flowParseGenericType();
      return;
    ***REMOVED***

    case tt.braceL:
      flowParseObjectType(false, false, false);
      return;

    case tt.braceBarL:
      flowParseObjectType(false, true, false);
      return;

    case tt.bracketL:
      flowParseTupleType();
      return;

    case tt.lessThan:
      flowParseTypeParameterDeclaration();
      expect(tt.parenL);
      flowParseFunctionTypeParams();
      expect(tt.parenR);
      expect(tt.arrow);
      flowParseType();
      return;

    case tt.parenL:
      next();

      // Check to see if this is actually a grouped type
      if (!match(tt.parenR) && !match(tt.ellipsis)) {
        if (match(tt.name)) {
          const token = lookaheadType();
          isGroupedType = token !== tt.question && token !== tt.colon;
        ***REMOVED*** else {
          isGroupedType = true;
        ***REMOVED***
      ***REMOVED***

      if (isGroupedType) {
        state.noAnonFunctionType = false;
        flowParseType();
        state.noAnonFunctionType = oldNoAnonFunctionType;

        // A `,` or a `) =>` means this is an anonymous function type
        if (
          state.noAnonFunctionType ||
          !(match(tt.comma) || (match(tt.parenR) && lookaheadType() === tt.arrow))
        ) {
          expect(tt.parenR);
          return;
        ***REMOVED*** else {
          // Eat a comma if there is one
          eat(tt.comma);
        ***REMOVED***
      ***REMOVED***

      flowParseFunctionTypeParams();

      expect(tt.parenR);
      expect(tt.arrow);
      flowParseType();
      return;

    case tt.minus:
      next();
      parseLiteral();
      return;

    case tt.string:
    case tt.num:
    case tt._true:
    case tt._false:
    case tt._null:
    case tt._this:
    case tt._void:
    case tt.star:
      next();
      return;

    default:
      if (state.type === tt._typeof) {
        flowParseTypeofType();
        return;
      ***REMOVED*** else if (state.type & TokenType.IS_KEYWORD) {
        next();
        state.tokens[state.tokens.length - 1].type = tt.name;
        return;
      ***REMOVED***
  ***REMOVED***

  unexpected();
***REMOVED***

function flowParsePostfixType() {
  flowParsePrimaryType();
  while (!canInsertSemicolon() && (match(tt.bracketL) || match(tt.questionDot))) {
    eat(tt.questionDot);
    expect(tt.bracketL);
    if (eat(tt.bracketR)) {
      // Array type
    ***REMOVED*** else {
      // Indexed access type
      flowParseType();
      expect(tt.bracketR);
    ***REMOVED***
  ***REMOVED***
***REMOVED***

function flowParsePrefixType() {
  if (eat(tt.question)) {
    flowParsePrefixType();
  ***REMOVED*** else {
    flowParsePostfixType();
  ***REMOVED***
***REMOVED***

function flowParseAnonFunctionWithoutParens() {
  flowParsePrefixType();
  if (!state.noAnonFunctionType && eat(tt.arrow)) {
    flowParseType();
  ***REMOVED***
***REMOVED***

function flowParseIntersectionType() {
  eat(tt.bitwiseAND);
  flowParseAnonFunctionWithoutParens();
  while (eat(tt.bitwiseAND)) {
    flowParseAnonFunctionWithoutParens();
  ***REMOVED***
***REMOVED***

function flowParseUnionType() {
  eat(tt.bitwiseOR);
  flowParseIntersectionType();
  while (eat(tt.bitwiseOR)) {
    flowParseIntersectionType();
  ***REMOVED***
***REMOVED***

function flowParseType() {
  flowParseUnionType();
***REMOVED***

export function flowParseTypeAnnotation() {
  flowParseTypeInitialiser();
***REMOVED***

function flowParseTypeAnnotatableIdentifier() {
  parseIdentifier();
  if (match(tt.colon)) {
    flowParseTypeAnnotation();
  ***REMOVED***
***REMOVED***

export function flowParseVariance() {
  if (match(tt.plus) || match(tt.minus)) {
    next();
    state.tokens[state.tokens.length - 1].isType = true;
  ***REMOVED***
***REMOVED***

// ==================================
// Overrides
// ==================================

export function flowParseFunctionBodyAndFinish(funcContextId) {
  // For arrow functions, `parseArrow` handles the return type itself.
  if (match(tt.colon)) {
    flowParseTypeAndPredicateInitialiser();
  ***REMOVED***

  parseFunctionBody(false, funcContextId);
***REMOVED***

export function flowParseSubscript(
  startTokenIndex,
  noCalls,
  stopState,
) {
  if (match(tt.questionDot) && lookaheadType() === tt.lessThan) {
    if (noCalls) {
      stopState.stop = true;
      return;
    ***REMOVED***
    next();
    flowParseTypeParameterInstantiation();
    expect(tt.parenL);
    parseCallExpressionArguments();
    return;
  ***REMOVED*** else if (!noCalls && match(tt.lessThan)) {
    const snapshot = state.snapshot();
    flowParseTypeParameterInstantiation();
    expect(tt.parenL);
    parseCallExpressionArguments();
    if (state.error) {
      state.restoreFromSnapshot(snapshot);
    ***REMOVED*** else {
      return;
    ***REMOVED***
  ***REMOVED***
  baseParseSubscript(startTokenIndex, noCalls, stopState);
***REMOVED***

export function flowStartParseNewArguments() {
  if (match(tt.lessThan)) {
    const snapshot = state.snapshot();
    flowParseTypeParameterInstantiation();
    if (state.error) {
      state.restoreFromSnapshot(snapshot);
    ***REMOVED***
  ***REMOVED***
***REMOVED***

// interfaces
export function flowTryParseStatement() {
  if (match(tt.name) && state.contextualKeyword === ContextualKeyword._interface) {
    const oldIsType = pushTypeContext(0);
    next();
    flowParseInterface();
    popTypeContext(oldIsType);
    return true;
  ***REMOVED*** else if (isContextual(ContextualKeyword._enum)) {
    flowParseEnumDeclaration();
    return true;
  ***REMOVED***
  return false;
***REMOVED***

export function flowTryParseExportDefaultExpression() {
  if (isContextual(ContextualKeyword._enum)) {
    flowParseEnumDeclaration();
    return true;
  ***REMOVED***
  return false;
***REMOVED***

// declares, interfaces and type aliases
export function flowParseIdentifierStatement(contextualKeyword) {
  if (contextualKeyword === ContextualKeyword._declare) {
    if (
      match(tt._class) ||
      match(tt.name) ||
      match(tt._function) ||
      match(tt._var) ||
      match(tt._export)
    ) {
      const oldIsType = pushTypeContext(1);
      flowParseDeclare();
      popTypeContext(oldIsType);
    ***REMOVED***
  ***REMOVED*** else if (match(tt.name)) {
    if (contextualKeyword === ContextualKeyword._interface) {
      const oldIsType = pushTypeContext(1);
      flowParseInterface();
      popTypeContext(oldIsType);
    ***REMOVED*** else if (contextualKeyword === ContextualKeyword._type) {
      const oldIsType = pushTypeContext(1);
      flowParseTypeAlias();
      popTypeContext(oldIsType);
    ***REMOVED*** else if (contextualKeyword === ContextualKeyword._opaque) {
      const oldIsType = pushTypeContext(1);
      flowParseOpaqueType(false);
      popTypeContext(oldIsType);
    ***REMOVED***
  ***REMOVED***
  semicolon();
***REMOVED***

// export type
export function flowShouldParseExportDeclaration() {
  return (
    isContextual(ContextualKeyword._type) ||
    isContextual(ContextualKeyword._interface) ||
    isContextual(ContextualKeyword._opaque) ||
    isContextual(ContextualKeyword._enum)
  );
***REMOVED***

export function flowShouldDisallowExportDefaultSpecifier() {
  return (
    match(tt.name) &&
    (state.contextualKeyword === ContextualKeyword._type ||
      state.contextualKeyword === ContextualKeyword._interface ||
      state.contextualKeyword === ContextualKeyword._opaque ||
      state.contextualKeyword === ContextualKeyword._enum)
  );
***REMOVED***

export function flowParseExportDeclaration() {
  if (isContextual(ContextualKeyword._type)) {
    const oldIsType = pushTypeContext(1);
    next();

    if (match(tt.braceL)) {
      // export type { foo, bar ***REMOVED***;
      parseExportSpecifiers();
      parseExportFrom();
    ***REMOVED*** else {
      // export type Foo = Bar;
      flowParseTypeAlias();
    ***REMOVED***
    popTypeContext(oldIsType);
  ***REMOVED*** else if (isContextual(ContextualKeyword._opaque)) {
    const oldIsType = pushTypeContext(1);
    next();
    // export opaque type Foo = Bar;
    flowParseOpaqueType(false);
    popTypeContext(oldIsType);
  ***REMOVED*** else if (isContextual(ContextualKeyword._interface)) {
    const oldIsType = pushTypeContext(1);
    next();
    flowParseInterface();
    popTypeContext(oldIsType);
  ***REMOVED*** else {
    parseStatement(true);
  ***REMOVED***
***REMOVED***

export function flowShouldParseExportStar() {
  return match(tt.star) || (isContextual(ContextualKeyword._type) && lookaheadType() === tt.star);
***REMOVED***

export function flowParseExportStar() {
  if (eatContextual(ContextualKeyword._type)) {
    const oldIsType = pushTypeContext(2);
    baseParseExportStar();
    popTypeContext(oldIsType);
  ***REMOVED*** else {
    baseParseExportStar();
  ***REMOVED***
***REMOVED***

// parse a the super class type parameters and implements
export function flowAfterParseClassSuper(hasSuper) {
  if (hasSuper && match(tt.lessThan)) {
    flowParseTypeParameterInstantiation();
  ***REMOVED***
  if (isContextual(ContextualKeyword._implements)) {
    const oldIsType = pushTypeContext(0);
    next();
    state.tokens[state.tokens.length - 1].type = tt._implements;
    do {
      flowParseRestrictedIdentifier();
      if (match(tt.lessThan)) {
        flowParseTypeParameterInstantiation();
      ***REMOVED***
    ***REMOVED*** while (eat(tt.comma));
    popTypeContext(oldIsType);
  ***REMOVED***
***REMOVED***

// parse type parameters for object method shorthand
export function flowStartParseObjPropValue() {
  // method shorthand
  if (match(tt.lessThan)) {
    flowParseTypeParameterDeclaration();
    if (!match(tt.parenL)) unexpected();
  ***REMOVED***
***REMOVED***

export function flowParseAssignableListItemTypes() {
  const oldIsType = pushTypeContext(0);
  eat(tt.question);
  if (match(tt.colon)) {
    flowParseTypeAnnotation();
  ***REMOVED***
  popTypeContext(oldIsType);
***REMOVED***

// parse typeof and type imports
export function flowStartParseImportSpecifiers() {
  if (match(tt._typeof) || isContextual(ContextualKeyword._type)) {
    const lh = lookaheadTypeAndKeyword();
    if (isMaybeDefaultImport(lh) || lh.type === tt.braceL || lh.type === tt.star) {
      next();
    ***REMOVED***
  ***REMOVED***
***REMOVED***

// parse import-type/typeof shorthand
export function flowParseImportSpecifier() {
  const isTypeKeyword =
    state.contextualKeyword === ContextualKeyword._type || state.type === tt._typeof;
  if (isTypeKeyword) {
    next();
  ***REMOVED*** else {
    parseIdentifier();
  ***REMOVED***

  if (isContextual(ContextualKeyword._as) && !isLookaheadContextual(ContextualKeyword._as)) {
    parseIdentifier();
    if (isTypeKeyword && !match(tt.name) && !(state.type & TokenType.IS_KEYWORD)) {
      // `import {type as ,` or `import {type as ***REMOVED***`
    ***REMOVED*** else {
      // `import {type as foo`
      parseIdentifier();
    ***REMOVED***
  ***REMOVED*** else {
    if (isTypeKeyword && (match(tt.name) || !!(state.type & TokenType.IS_KEYWORD))) {
      // `import {type foo`
      parseIdentifier();
    ***REMOVED***
    if (eatContextual(ContextualKeyword._as)) {
      parseIdentifier();
    ***REMOVED***
  ***REMOVED***
***REMOVED***

// parse function type parameters - function foo<T>() {***REMOVED***
export function flowStartParseFunctionParams() {
  // Originally this checked if the method is a getter/setter, but if it was, we'd crash soon
  // anyway, so don't try to propagate that information.
  if (match(tt.lessThan)) {
    const oldIsType = pushTypeContext(0);
    flowParseTypeParameterDeclaration();
    popTypeContext(oldIsType);
  ***REMOVED***
***REMOVED***

// parse flow type annotations on variable declarator heads - let foo: string = bar
export function flowAfterParseVarHead() {
  if (match(tt.colon)) {
    flowParseTypeAnnotation();
  ***REMOVED***
***REMOVED***

// parse the return type of an async arrow function - let foo = (async (): number => {***REMOVED***);
export function flowStartParseAsyncArrowFromCallExpression() {
  if (match(tt.colon)) {
    const oldNoAnonFunctionType = state.noAnonFunctionType;
    state.noAnonFunctionType = true;
    flowParseTypeAnnotation();
    state.noAnonFunctionType = oldNoAnonFunctionType;
  ***REMOVED***
***REMOVED***

// We need to support type parameter declarations for arrow functions. This
// is tricky. There are three situations we need to handle
//
// 1. This is either JSX or an arrow function. We'll try JSX first. If that
//    fails, we'll try an arrow function. If that fails, we'll throw the JSX
//    error.
// 2. This is an arrow function. We'll parse the type parameter declaration,
//    parse the rest, make sure the rest is an arrow function, and go from
//    there
// 3. This is neither. Just call the super method
export function flowParseMaybeAssign(noIn, isWithinParens) {
  if (match(tt.lessThan)) {
    const snapshot = state.snapshot();
    let wasArrow = baseParseMaybeAssign(noIn, isWithinParens);
    if (state.error) {
      state.restoreFromSnapshot(snapshot);
      state.type = tt.typeParameterStart;
    ***REMOVED*** else {
      return wasArrow;
    ***REMOVED***

    const oldIsType = pushTypeContext(0);
    flowParseTypeParameterDeclaration();
    popTypeContext(oldIsType);
    wasArrow = baseParseMaybeAssign(noIn, isWithinParens);
    if (wasArrow) {
      return true;
    ***REMOVED***
    unexpected();
  ***REMOVED***

  return baseParseMaybeAssign(noIn, isWithinParens);
***REMOVED***

// handle return types for arrow functions
export function flowParseArrow() {
  if (match(tt.colon)) {
    const oldIsType = pushTypeContext(0);
    const snapshot = state.snapshot();

    const oldNoAnonFunctionType = state.noAnonFunctionType;
    state.noAnonFunctionType = true;
    flowParseTypeAndPredicateInitialiser();
    state.noAnonFunctionType = oldNoAnonFunctionType;

    if (canInsertSemicolon()) unexpected();
    if (!match(tt.arrow)) unexpected();

    if (state.error) {
      state.restoreFromSnapshot(snapshot);
    ***REMOVED***
    popTypeContext(oldIsType);
  ***REMOVED***
  return eat(tt.arrow);
***REMOVED***

export function flowParseSubscripts(startTokenIndex, noCalls = false) {
  if (
    state.tokens[state.tokens.length - 1].contextualKeyword === ContextualKeyword._async &&
    match(tt.lessThan)
  ) {
    const snapshot = state.snapshot();
    const wasArrow = parseAsyncArrowWithTypeParameters();
    if (wasArrow && !state.error) {
      return;
    ***REMOVED***
    state.restoreFromSnapshot(snapshot);
  ***REMOVED***

  baseParseSubscripts(startTokenIndex, noCalls);
***REMOVED***

// Returns true if there was an arrow function here.
function parseAsyncArrowWithTypeParameters() {
  state.scopeDepth++;
  const startTokenIndex = state.tokens.length;
  parseFunctionParams();
  if (!parseArrow()) {
    return false;
  ***REMOVED***
  parseArrowExpression(startTokenIndex);
  return true;
***REMOVED***

function flowParseEnumDeclaration() {
  expectContextual(ContextualKeyword._enum);
  state.tokens[state.tokens.length - 1].type = tt._enum;
  parseIdentifier();
  flowParseEnumBody();
***REMOVED***

function flowParseEnumBody() {
  if (eatContextual(ContextualKeyword._of)) {
    next();
  ***REMOVED***
  expect(tt.braceL);
  flowParseEnumMembers();
  expect(tt.braceR);
***REMOVED***

function flowParseEnumMembers() {
  while (!match(tt.braceR) && !state.error) {
    if (eat(tt.ellipsis)) {
      break;
    ***REMOVED***
    flowParseEnumMember();
    if (!match(tt.braceR)) {
      expect(tt.comma);
    ***REMOVED***
  ***REMOVED***
***REMOVED***

function flowParseEnumMember() {
  parseIdentifier();
  if (eat(tt.eq)) {
    // Flow enum values are always just one token (a string, number, or boolean literal).
    next();
  ***REMOVED***
***REMOVED***
