// @babel/plugin-bugfix-safari-class-field-initializer-scope/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true ***REMOVED***);

var helperPluginUtils = require('@babel/helper-plugin-utils');
var core = require('@babel/core');

function needsWrapping(node) {
  if (core.types.isLiteral(node) && !core.types.isTemplateLiteral(node)) {
    return false;
  ***REMOVED***
  if (core.types.isCallExpression(node) || core.types.isOptionalCallExpression(node) || core.types.isNewExpression(node)) {
    return needsWrapping(node.callee) || node.arguments.some(needsWrapping);
  ***REMOVED***
  if (core.types.isTemplateLiteral(node)) {
    return node.expressions.some(needsWrapping);
  ***REMOVED***
  if (core.types.isTaggedTemplateExpression(node)) {
    return needsWrapping(node.tag) || needsWrapping(node.quasi);
  ***REMOVED***
  if (core.types.isArrayExpression(node)) {
    return node.elements.some(needsWrapping);
  ***REMOVED***
  if (core.types.isObjectExpression(node)) {
    return node.properties.some(prop => {
      if (core.types.isObjectProperty(prop)) {
        return needsWrapping(prop.value) || prop.computed && needsWrapping(prop.key);
      ***REMOVED***
      if (core.types.isObjectMethod(prop)) {
        return false;
      ***REMOVED***
      return false;
    ***REMOVED***);
  ***REMOVED***
  if (core.types.isMemberExpression(node) || core.types.isOptionalMemberExpression(node)) {
    return needsWrapping(node.object) || node.computed && needsWrapping(node.property);
  ***REMOVED***
  if (core.types.isFunctionExpression(node) || core.types.isArrowFunctionExpression(node) || core.types.isClassExpression(node)) {
    return false;
  ***REMOVED***
  if (core.types.isThisExpression(node)) {
    return false;
  ***REMOVED***
  if (core.types.isSequenceExpression(node)) {
    return node.expressions.some(needsWrapping);
  ***REMOVED***
  return true;
***REMOVED***
function wrapInitializer(path) {
  const {
    value
  ***REMOVED*** = path.node;
  if (value && needsWrapping(value)) {
    path.node.value = core.types.callExpression(core.types.arrowFunctionExpression([], value), []);
  ***REMOVED***
***REMOVED***
var index = helperPluginUtils.declare(api => {
  api.assertVersion("^7.16.0");
  return {
    name: "plugin-bugfix-safari-class-field-initializer-scope",
    visitor: {
      ClassProperty(path) {
        wrapInitializer(path);
      ***REMOVED***,
      ClassPrivateProperty(path) {
        wrapInitializer(path);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;
***REMOVED***);

exports.default = index;
