// @babel/core/lib/config/item.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.createConfigItem = createConfigItem;
exports.createItemFromDescriptor = createItemFromDescriptor;
exports.getItemDescriptor = getItemDescriptor;
function _path() {
  const data = require("path");
  _path = function () {
    return data;
  ***REMOVED***;
  return data;
***REMOVED***
var _configDescriptors = require("./config-descriptors.js");
function createItemFromDescriptor(desc) {
  return new ConfigItem(desc);
***REMOVED***
function* createConfigItem(value, {
  dirname = ".",
  type
***REMOVED*** = {***REMOVED***) {
  const descriptor = yield* (0, _configDescriptors.createDescriptor)(value, _path().resolve(dirname), {
    type,
    alias: "programmatic item"
  ***REMOVED***);
  return createItemFromDescriptor(descriptor);
***REMOVED***
const CONFIG_ITEM_BRAND = Symbol.for("@babel/core@7 - ConfigItem");
function getItemDescriptor(item) {
  if (item != null && item[CONFIG_ITEM_BRAND]) {
    return item._descriptor;
  ***REMOVED***
  return undefined;
***REMOVED***
class ConfigItem {
  constructor(descriptor) {
    this._descriptor = void 0;
    this[CONFIG_ITEM_BRAND] = true;
    this.value = void 0;
    this.options = void 0;
    this.dirname = void 0;
    this.name = void 0;
    this.file = void 0;
    this._descriptor = descriptor;
    Object.defineProperty(this, "_descriptor", {
      enumerable: false
    ***REMOVED***);
    Object.defineProperty(this, CONFIG_ITEM_BRAND, {
      enumerable: false
    ***REMOVED***);
    this.value = this._descriptor.value;
    this.options = this._descriptor.options;
    this.dirname = this._descriptor.dirname;
    this.name = this._descriptor.name;
    this.file = this._descriptor.file ? {
      request: this._descriptor.file.request,
      resolved: this._descriptor.file.resolved
    ***REMOVED*** : undefined;
    Object.freeze(this);
  ***REMOVED***
***REMOVED***
Object.freeze(ConfigItem.prototype);
0 && 0;
