// @babel/core/lib/config/printer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.ConfigPrinter = exports.ChainFormatter = void 0;
function _gensync() {
  const data = require("gensync");
  _gensync = function () {
    return data;
  ***REMOVED***;
  return data;
***REMOVED***
const ChainFormatter = exports.ChainFormatter = {
  Programmatic: 0,
  Config: 1
***REMOVED***;
const Formatter = {
  title(type, callerName, filepath) {
    let title = "";
    if (type === ChainFormatter.Programmatic) {
      title = "programmatic options";
      if (callerName) {
        title += " from " + callerName;
      ***REMOVED***
    ***REMOVED*** else {
      title = "config " + filepath;
    ***REMOVED***
    return title;
  ***REMOVED***,
  loc(index, envName) {
    let loc = "";
    if (index != null) {
      loc += `.overrides[${index***REMOVED***]`;
    ***REMOVED***
    if (envName != null) {
      loc += `.env["${envName***REMOVED***"]`;
    ***REMOVED***
    return loc;
  ***REMOVED***,
  *optionsAndDescriptors(opt) {
    const content = Object.assign({***REMOVED***, opt.options);
    delete content.overrides;
    delete content.env;
    const pluginDescriptors = [...(yield* opt.plugins())];
    if (pluginDescriptors.length) {
      content.plugins = pluginDescriptors.map(d => descriptorToConfig(d));
    ***REMOVED***
    const presetDescriptors = [...(yield* opt.presets())];
    if (presetDescriptors.length) {
      content.presets = [...presetDescriptors].map(d => descriptorToConfig(d));
    ***REMOVED***
    return JSON.stringify(content, undefined, 2);
  ***REMOVED***
***REMOVED***;
function descriptorToConfig(d) {
  var _d$file;
  let name = (_d$file = d.file) == null ? void 0 : _d$file.request;
  if (name == null) {
    if (typeof d.value === "object") {
      name = d.value;
    ***REMOVED*** else if (typeof d.value === "function") {
      name = `[Function: ${d.value.toString().slice(0, 50)***REMOVED*** ... ]`;
    ***REMOVED***
  ***REMOVED***
  if (name == null) {
    name = "[Unknown]";
  ***REMOVED***
  if (d.options === undefined) {
    return name;
  ***REMOVED*** else if (d.name == null) {
    return [name, d.options];
  ***REMOVED*** else {
    return [name, d.options, d.name];
  ***REMOVED***
***REMOVED***
class ConfigPrinter {
  constructor() {
    this._stack = [];
  ***REMOVED***
  configure(enabled, type, {
    callerName,
    filepath
  ***REMOVED***) {
    if (!enabled) return () => {***REMOVED***;
    return (content, index, envName) => {
      this._stack.push({
        type,
        callerName,
        filepath,
        content,
        index,
        envName
      ***REMOVED***);
    ***REMOVED***;
  ***REMOVED***
  static *format(config) {
    let title = Formatter.title(config.type, config.callerName, config.filepath);
    const loc = Formatter.loc(config.index, config.envName);
    if (loc) title += ` ${loc***REMOVED***`;
    const content = yield* Formatter.optionsAndDescriptors(config.content);
    return `${title***REMOVED***\n${content***REMOVED***`;
  ***REMOVED***
  *output() {
    if (this._stack.length === 0) return "";
    const configs = yield* _gensync().all(this._stack.map(s => ConfigPrinter.format(s)));
    return configs.join("\n\n");
  ***REMOVED***
***REMOVED***
exports.ConfigPrinter = ConfigPrinter;
0 && 0;
