// @babel/core/lib/config/resolve-targets.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.resolveBrowserslistConfigFile = resolveBrowserslistConfigFile;
exports.resolveTargets = resolveTargets;
function _path() {
  const data = require("path");
  _path = function () {
    return data;
  ***REMOVED***;
  return data;
***REMOVED***
function _helperCompilationTargets() {
  const data = require("@babel/helper-compilation-targets");
  _helperCompilationTargets = function () {
    return data;
  ***REMOVED***;
  return data;
***REMOVED***
({***REMOVED***);
function resolveBrowserslistConfigFile(browserslistConfigFile, configFileDir) {
  return _path().resolve(configFileDir, browserslistConfigFile);
***REMOVED***
function resolveTargets(options, root) {
  const optTargets = options.targets;
  let targets;
  if (typeof optTargets === "string" || Array.isArray(optTargets)) {
    targets = {
      browsers: optTargets
    ***REMOVED***;
  ***REMOVED*** else if (optTargets) {
    if ("esmodules" in optTargets) {
      targets = Object.assign({***REMOVED***, optTargets, {
        esmodules: "intersect"
      ***REMOVED***);
    ***REMOVED*** else {
      targets = optTargets;
    ***REMOVED***
  ***REMOVED***
  const {
    browserslistConfigFile
  ***REMOVED*** = options;
  let configFile;
  let ignoreBrowserslistConfig = false;
  if (typeof browserslistConfigFile === "string") {
    configFile = browserslistConfigFile;
  ***REMOVED*** else {
    ignoreBrowserslistConfig = browserslistConfigFile === false;
  ***REMOVED***
  return (0, _helperCompilationTargets().default)(targets, {
    ignoreBrowserslistConfig,
    configFile,
    configPath: root,
    browserslistEnv: options.browserslistEnv
  ***REMOVED***);
***REMOVED***
0 && 0;
