// @babel/core/lib/config/resolve-targets-browser.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.resolveBrowserslistConfigFile = resolveBrowserslistConfigFile;
exports.resolveTargets = resolveTargets;
function _helperCompilationTargets() {
  const data = require("@babel/helper-compilation-targets");
  _helperCompilationTargets = function () {
    return data;
  ***REMOVED***;
  return data;
***REMOVED***
function resolveBrowserslistConfigFile(browserslistConfigFile, configFilePath) {
  return undefined;
***REMOVED***
function resolveTargets(options, root) {
  const optTargets = options.targets;
  let targets;
  if (typeof optTargets === "string" || Array.isArray(optTargets)) {
    targets = {
      browsers: optTargets
    ***REMOVED***;
  ***REMOVED*** else if (optTargets) {
    if ("esmodules" in optTargets) {
      targets = Object.assign({***REMOVED***, optTargets, {
        esmodules: "intersect"
      ***REMOVED***);
    ***REMOVED*** else {
      targets = optTargets;
    ***REMOVED***
  ***REMOVED***
  return (0, _helperCompilationTargets().default)(targets, {
    ignoreBrowserslistConfig: true,
    browserslistEnv: options.browserslistEnv
  ***REMOVED***);
***REMOVED***
0 && 0;
