// eslint/lib/config/rule-validator.js
/**
 * @fileoverview Rule Validator
 * @author Nicholas C. Zakas
 */

"use strict";

//-----------------------------------------------------------------------------
// Requirements
//-----------------------------------------------------------------------------

const ajv = require("../shared/ajv")();

//-----------------------------------------------------------------------------
// Helpers
//-----------------------------------------------------------------------------

/**
 * Finds a rule with the given ID in the given config.
 * @param {string***REMOVED*** ruleId The ID of the rule to find.
 * @param {Object***REMOVED*** config The config to search in.
 * @returns {{create: Function, schema: (Array|null)***REMOVED******REMOVED*** THe rule object.
 */
function findRuleDefinition(ruleId, config) {
    const ruleIdParts = ruleId.split("/");
    let pluginName, ruleName;

    // built-in rule
    if (ruleIdParts.length === 1) {
        pluginName = "@";
        ruleName = ruleIdParts[0];
    ***REMOVED*** else {
        ruleName = ruleIdParts.pop();
        pluginName = ruleIdParts.join("/");
    ***REMOVED***

    if (!config.plugins || !config.plugins[pluginName]) {
        throw new TypeError(`Key "rules": Key "${ruleId***REMOVED***": Could not find plugin "${pluginName***REMOVED***".`);
    ***REMOVED***

    if (!config.plugins[pluginName].rules || !config.plugins[pluginName].rules[ruleName]) {
        throw new TypeError(`Key "rules": Key "${ruleId***REMOVED***": Could not find "${ruleName***REMOVED***" in plugin "${pluginName***REMOVED***".`);
    ***REMOVED***

    return config.plugins[pluginName].rules[ruleName];

***REMOVED***

/**
 * Gets a complete options schema for a rule.
 * @param {{create: Function, schema: (Array|null)***REMOVED******REMOVED*** rule A new-style rule object
 * @returns {Object***REMOVED*** JSON Schema for the rule's options.
 */
function getRuleOptionsSchema(rule) {

    if (!rule) {
        return null;
    ***REMOVED***

    const schema = rule.schema || rule.meta && rule.meta.schema;

    if (Array.isArray(schema)) {
        if (schema.length) {
            return {
                type: "array",
                items: schema,
                minItems: 0,
                maxItems: schema.length
            ***REMOVED***;
        ***REMOVED***
        return {
            type: "array",
            minItems: 0,
            maxItems: 0
        ***REMOVED***;

    ***REMOVED***

    // Given a full schema, leave it alone
    return schema || null;
***REMOVED***

//-----------------------------------------------------------------------------
// Exports
//-----------------------------------------------------------------------------

/**
 * Implements validation functionality for the rules portion of a config.
 */
class RuleValidator {

    /**
     * Creates a new instance.
     */
    constructor() {

        /**
         * A collection of compiled validators for rules that have already
         * been validated.
         * @type {WeakMap***REMOVED***
         * @property validators
         */
        this.validators = new WeakMap();
    ***REMOVED***

    /**
     * Validates all of the rule configurations in a config against each
     * rule's schema.
     * @param {Object***REMOVED*** config The full config to validate. This object must
     *      contain both the rules section and the plugins section.
     * @returns {void***REMOVED***
     * @throws {Error***REMOVED*** If a rule's configuration does not match its schema.
     */
    validate(config) {

        if (!config.rules) {
            return;
        ***REMOVED***

        for (const [ruleId, ruleOptions] of Object.entries(config.rules)) {

            // check for edge case
            if (ruleId === "__proto__") {
                continue;
            ***REMOVED***

            /*
             * If a rule is disabled, we don't do any validation. This allows
             * users to safely set any value to 0 or "off" without worrying
             * that it will cause a validation error.
             *
             * Note: ruleOptions is always an array at this point because
             * this validation occurs after FlatConfigArray has merged and
             * normalized values.
             */
            if (ruleOptions[0] === 0) {
                continue;
            ***REMOVED***

            const rule = findRuleDefinition(ruleId, config);

            // Precompile and cache validator the first time
            if (!this.validators.has(rule)) {
                const schema = getRuleOptionsSchema(rule);

                if (schema) {
                    this.validators.set(rule, ajv.compile(schema));
                ***REMOVED***
            ***REMOVED***

            const validateRule = this.validators.get(rule);

            if (validateRule) {

                validateRule(ruleOptions.slice(1));

                if (validateRule.errors) {
                    throw new Error(`Key "rules": Key "${ruleId***REMOVED***": ${
                        validateRule.errors.map(
                            error => `\tValue ${JSON.stringify(error.data)***REMOVED*** ${error.message***REMOVED***.\n`
                        ).join("")
                    ***REMOVED***`);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
***REMOVED***

exports.RuleValidator = RuleValidator;
