// handle-thing/lib/queue.js
function Queue () {
  this.head = new Item('head', null)
***REMOVED***
module.exports = Queue

Queue.prototype.append = function append (kind, value) {
  var item = new Item(kind, value)
  this.head.prepend(item)
  return item
***REMOVED***

Queue.prototype.isEmpty = function isEmpty () {
  return this.head.prev === this.head
***REMOVED***

Queue.prototype.first = function first () {
  return this.head.next
***REMOVED***

function Item (kind, value) {
  this.prev = this
  this.next = this
  this.kind = kind
  this.value = value
***REMOVED***

Item.prototype.prepend = function prepend (other) {
  other.prev = this.prev
  other.next = this
  other.prev.next = other
  other.next.prev = other
***REMOVED***

Item.prototype.dequeue = function dequeue () {
  var prev = this.prev
  var next = this.next

  prev.next = next
  next.prev = prev
  this.prev = this
  this.next = this

  return this.value
***REMOVED***

Item.prototype.isEmpty = function isEmpty () {
  return this.prev === this
***REMOVED***
