// webpack/lib/wasm-sync/WebAssemblyGenerator.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const t = require("@webassemblyjs/ast");
const { moduleContextFromModuleAST ***REMOVED*** = require("@webassemblyjs/ast");
const { editWithAST, addWithAST ***REMOVED*** = require("@webassemblyjs/wasm-edit");
const { decode ***REMOVED*** = require("@webassemblyjs/wasm-parser");
const { RawSource ***REMOVED*** = require("webpack-sources");
const Generator = require("../Generator");
const { WEBASSEMBLY_TYPES ***REMOVED*** = require("../ModuleSourceTypesConstants");
const WebAssemblyUtils = require("./WebAssemblyUtils");

const WebAssemblyExportImportedDependency = require("../dependencies/WebAssemblyExportImportedDependency");

/** @typedef {import("webpack-sources").Source***REMOVED*** Source */
/** @typedef {import("../DependencyTemplates")***REMOVED*** DependencyTemplates */
/** @typedef {import("../Generator").GenerateContext***REMOVED*** GenerateContext */
/** @typedef {import("../Module")***REMOVED*** Module */
/** @typedef {import("../Module").SourceTypes***REMOVED*** SourceTypes */
/** @typedef {import("../ModuleGraph")***REMOVED*** ModuleGraph */
/** @typedef {import("../NormalModule")***REMOVED*** NormalModule */
/** @typedef {import("../RuntimeTemplate")***REMOVED*** RuntimeTemplate */
/** @typedef {import("../util/runtime").RuntimeSpec***REMOVED*** RuntimeSpec */
/** @typedef {import("./WebAssemblyUtils").UsedWasmDependency***REMOVED*** UsedWasmDependency */
/** @typedef {import("@webassemblyjs/ast").Instruction***REMOVED*** Instruction */
/** @typedef {import("@webassemblyjs/ast").ModuleImport***REMOVED*** ModuleImport */
/** @typedef {import("@webassemblyjs/ast").ModuleExport***REMOVED*** ModuleExport */
/** @typedef {import("@webassemblyjs/ast").Global***REMOVED*** Global */
/** @typedef {import("@webassemblyjs/ast").AST***REMOVED*** AST */
/** @typedef {import("@webassemblyjs/ast").GlobalType***REMOVED*** GlobalType */
/**
 * @template T
 * @typedef {import("@webassemblyjs/ast").NodePath<T>***REMOVED*** NodePath
 */

/**
 * @typedef {(buf: ArrayBuffer) => ArrayBuffer***REMOVED*** ArrayBufferTransform
 */

/**
 * @template T
 * @param {((prev: ArrayBuffer) => ArrayBuffer)[]***REMOVED*** fns transforms
 * @returns {ArrayBufferTransform***REMOVED*** composed transform
 */
const compose = (...fns) =>
	fns.reduce(
		(prevFn, nextFn) => value => nextFn(prevFn(value)),
		value => value
	);

/**
 * Removes the start instruction
 * @param {object***REMOVED*** state state
 * @param {AST***REMOVED*** state.ast Module's ast
 * @returns {ArrayBufferTransform***REMOVED*** transform
 */
const removeStartFunc = state => bin =>
	editWithAST(state.ast, bin, {
		Start(path) {
			path.remove();
		***REMOVED***
	***REMOVED***);

/**
 * Get imported globals
 * @param {AST***REMOVED*** ast Module's AST
 * @returns {t.ModuleImport[]***REMOVED*** - nodes
 */
const getImportedGlobals = ast => {
	/** @type {t.ModuleImport[]***REMOVED*** */
	const importedGlobals = [];

	t.traverse(ast, {
		ModuleImport({ node ***REMOVED***) {
			if (t.isGlobalType(node.descr)) {
				importedGlobals.push(node);
			***REMOVED***
		***REMOVED***
	***REMOVED***);

	return importedGlobals;
***REMOVED***;

/**
 * Get the count for imported func
 * @param {AST***REMOVED*** ast Module's AST
 * @returns {number***REMOVED*** - count
 */
const getCountImportedFunc = ast => {
	let count = 0;

	t.traverse(ast, {
		ModuleImport({ node ***REMOVED***) {
			if (t.isFuncImportDescr(node.descr)) {
				count++;
			***REMOVED***
		***REMOVED***
	***REMOVED***);

	return count;
***REMOVED***;

/**
 * Get next type index
 * @param {AST***REMOVED*** ast Module's AST
 * @returns {t.Index***REMOVED*** - index
 */
const getNextTypeIndex = ast => {
	const typeSectionMetadata = t.getSectionMetadata(ast, "type");

	if (typeSectionMetadata === undefined) {
		return t.indexLiteral(0);
	***REMOVED***

	return t.indexLiteral(typeSectionMetadata.vectorOfSize.value);
***REMOVED***;

/**
 * Get next func index
 * The Func section metadata provide information for implemented funcs
 * in order to have the correct index we shift the index by number of external
 * functions.
 * @param {AST***REMOVED*** ast Module's AST
 * @param {number***REMOVED*** countImportedFunc number of imported funcs
 * @returns {t.Index***REMOVED*** - index
 */
const getNextFuncIndex = (ast, countImportedFunc) => {
	const funcSectionMetadata = t.getSectionMetadata(ast, "func");

	if (funcSectionMetadata === undefined) {
		return t.indexLiteral(0 + countImportedFunc);
	***REMOVED***

	const vectorOfSize = funcSectionMetadata.vectorOfSize.value;

	return t.indexLiteral(vectorOfSize + countImportedFunc);
***REMOVED***;

/**
 * Creates an init instruction for a global type
 * @param {t.GlobalType***REMOVED*** globalType the global type
 * @returns {t.Instruction***REMOVED*** init expression
 */
const createDefaultInitForGlobal = globalType => {
	if (globalType.valtype[0] === "i") {
		// create NumberLiteral global initializer
		return t.objectInstruction("const", globalType.valtype, [
			t.numberLiteralFromRaw(66)
		]);
	***REMOVED*** else if (globalType.valtype[0] === "f") {
		// create FloatLiteral global initializer
		return t.objectInstruction("const", globalType.valtype, [
			t.floatLiteral(66, false, false, "66")
		]);
	***REMOVED***
	throw new Error(`unknown type: ${globalType.valtype***REMOVED***`);
***REMOVED***;

/**
 * Rewrite the import globals:
 * - removes the ModuleImport instruction
 * - injects at the same offset a mutable global of the same type
 *
 * Since the imported globals are before the other global declarations, our
 * indices will be preserved.
 *
 * Note that globals will become mutable.
 * @param {object***REMOVED*** state transformation state
 * @param {AST***REMOVED*** state.ast Module's ast
 * @param {t.Instruction[]***REMOVED*** state.additionalInitCode list of addition instructions for the init function
 * @returns {ArrayBufferTransform***REMOVED*** transform
 */
const rewriteImportedGlobals = state => bin => {
	const additionalInitCode = state.additionalInitCode;
	/** @type {Array<t.Global>***REMOVED*** */
	const newGlobals = [];

	bin = editWithAST(state.ast, bin, {
		ModuleImport(path) {
			if (t.isGlobalType(path.node.descr)) {
				const globalType =
					/** @type {GlobalType***REMOVED*** */
					(path.node.descr);

				globalType.mutability = "var";

				const init = [
					createDefaultInitForGlobal(globalType),
					t.instruction("end")
				];

				newGlobals.push(t.global(globalType, init));

				path.remove();
			***REMOVED***
		***REMOVED***,

		// in order to preserve non-imported global's order we need to re-inject
		// those as well
		/**
		 * @param {NodePath<Global>***REMOVED*** path path
		 */
		Global(path) {
			const { node ***REMOVED*** = path;
			const [init] = node.init;

			if (init.id === "get_global") {
				node.globalType.mutability = "var";

				const initialGlobalIdx = init.args[0];

				node.init = [
					createDefaultInitForGlobal(node.globalType),
					t.instruction("end")
				];

				additionalInitCode.push(
					/**
					 * get_global in global initializer only works for imported globals.
					 * They have the same indices as the init params, so use the
					 * same index.
					 */
					t.instruction("get_local", [initialGlobalIdx]),
					t.instruction("set_global", [t.indexLiteral(newGlobals.length)])
				);
			***REMOVED***

			newGlobals.push(node);

			path.remove();
		***REMOVED***
	***REMOVED***);

	// Add global declaration instructions
	return addWithAST(state.ast, bin, newGlobals);
***REMOVED***;

/**
 * Rewrite the export names
 * @param {object***REMOVED*** state state
 * @param {AST***REMOVED*** state.ast Module's ast
 * @param {Module***REMOVED*** state.module Module
 * @param {ModuleGraph***REMOVED*** state.moduleGraph module graph
 * @param {Set<string>***REMOVED*** state.externalExports Module
 * @param {RuntimeSpec***REMOVED*** state.runtime runtime
 * @returns {ArrayBufferTransform***REMOVED*** transform
 */
const rewriteExportNames =
	({ ast, moduleGraph, module, externalExports, runtime ***REMOVED***) =>
	bin =>
		editWithAST(ast, bin, {
			/**
			 * @param {NodePath<ModuleExport>***REMOVED*** path path
			 */
			ModuleExport(path) {
				const isExternal = externalExports.has(path.node.name);
				if (isExternal) {
					path.remove();
					return;
				***REMOVED***
				const usedName = moduleGraph
					.getExportsInfo(module)
					.getUsedName(path.node.name, runtime);
				if (!usedName) {
					path.remove();
					return;
				***REMOVED***
				path.node.name = /** @type {string***REMOVED*** */ (usedName);
			***REMOVED***
		***REMOVED***);

/**
 * Mangle import names and modules
 * @param {object***REMOVED*** state state
 * @param {AST***REMOVED*** state.ast Module's ast
 * @param {Map<string, UsedWasmDependency>***REMOVED*** state.usedDependencyMap mappings to mangle names
 * @returns {ArrayBufferTransform***REMOVED*** transform
 */
const rewriteImports =
	({ ast, usedDependencyMap ***REMOVED***) =>
	bin =>
		editWithAST(ast, bin, {
			/**
			 * @param {NodePath<ModuleImport>***REMOVED*** path path
			 */
			ModuleImport(path) {
				const result = usedDependencyMap.get(
					`${path.node.module***REMOVED***:${path.node.name***REMOVED***`
				);

				if (result !== undefined) {
					path.node.module = result.module;
					path.node.name = result.name;
				***REMOVED***
			***REMOVED***
		***REMOVED***);

/**
 * Add an init function.
 *
 * The init function fills the globals given input arguments.
 * @param {object***REMOVED*** state transformation state
 * @param {AST***REMOVED*** state.ast Module's ast
 * @param {t.Identifier***REMOVED*** state.initFuncId identifier of the init function
 * @param {t.Index***REMOVED*** state.startAtFuncOffset index of the start function
 * @param {t.ModuleImport[]***REMOVED*** state.importedGlobals list of imported globals
 * @param {t.Instruction[]***REMOVED*** state.additionalInitCode list of addition instructions for the init function
 * @param {t.Index***REMOVED*** state.nextFuncIndex index of the next function
 * @param {t.Index***REMOVED*** state.nextTypeIndex index of the next type
 * @returns {ArrayBufferTransform***REMOVED*** transform
 */
const addInitFunction =
	({
		ast,
		initFuncId,
		startAtFuncOffset,
		importedGlobals,
		additionalInitCode,
		nextFuncIndex,
		nextTypeIndex
	***REMOVED***) =>
	bin => {
		const funcParams = importedGlobals.map(importedGlobal => {
			// used for debugging
			const id = t.identifier(
				`${importedGlobal.module***REMOVED***.${importedGlobal.name***REMOVED***`
			);

			return t.funcParam(
				/** @type {string***REMOVED*** */ (importedGlobal.descr.valtype),
				id
			);
		***REMOVED***);

		/** @type {Instruction[]***REMOVED*** */
		const funcBody = [];
		for (const [index, _importedGlobal] of importedGlobals.entries()) {
			const args = [t.indexLiteral(index)];
			const body = [
				t.instruction("get_local", args),
				t.instruction("set_global", args)
			];

			funcBody.push(...body);
		***REMOVED***

		if (typeof startAtFuncOffset === "number") {
			funcBody.push(
				t.callInstruction(t.numberLiteralFromRaw(startAtFuncOffset))
			);
		***REMOVED***

		for (const instr of additionalInitCode) {
			funcBody.push(instr);
		***REMOVED***

		funcBody.push(t.instruction("end"));

		/** @type {string[]***REMOVED*** */
		const funcResults = [];

		// Code section
		const funcSignature = t.signature(funcParams, funcResults);
		const func = t.func(initFuncId, funcSignature, funcBody);

		// Type section
		const functype = t.typeInstruction(undefined, funcSignature);

		// Func section
		const funcindex = t.indexInFuncSection(nextTypeIndex);

		// Export section
		const moduleExport = t.moduleExport(
			initFuncId.value,
			t.moduleExportDescr("Func", nextFuncIndex)
		);

		return addWithAST(ast, bin, [func, moduleExport, funcindex, functype]);
	***REMOVED***;

/**
 * Extract mangle mappings from module
 * @param {ModuleGraph***REMOVED*** moduleGraph module graph
 * @param {Module***REMOVED*** module current module
 * @param {boolean | undefined***REMOVED*** mangle mangle imports
 * @returns {Map<string, UsedWasmDependency>***REMOVED*** mappings to mangled names
 */
const getUsedDependencyMap = (moduleGraph, module, mangle) => {
	/** @type {Map<string, UsedWasmDependency>***REMOVED*** */
	const map = new Map();
	for (const usedDep of WebAssemblyUtils.getUsedDependencies(
		moduleGraph,
		module,
		mangle
	)) {
		const dep = usedDep.dependency;
		const request = dep.request;
		const exportName = dep.name;
		map.set(`${request***REMOVED***:${exportName***REMOVED***`, usedDep);
	***REMOVED***
	return map;
***REMOVED***;

/**
 * @typedef {object***REMOVED*** WebAssemblyGeneratorOptions
 * @property {boolean=***REMOVED*** mangleImports mangle imports
 */

class WebAssemblyGenerator extends Generator {
	/**
	 * @param {WebAssemblyGeneratorOptions***REMOVED*** options options
	 */
	constructor(options) {
		super();
		this.options = options;
	***REMOVED***

	/**
	 * @param {NormalModule***REMOVED*** module fresh module
	 * @returns {SourceTypes***REMOVED*** available types (do not mutate)
	 */
	getTypes(module) {
		return WEBASSEMBLY_TYPES;
	***REMOVED***

	/**
	 * @param {NormalModule***REMOVED*** module the module
	 * @param {string=***REMOVED*** type source type
	 * @returns {number***REMOVED*** estimate size of the module
	 */
	getSize(module, type) {
		const originalSource = module.originalSource();
		if (!originalSource) {
			return 0;
		***REMOVED***
		return originalSource.size();
	***REMOVED***

	/**
	 * @param {NormalModule***REMOVED*** module module for which the code should be generated
	 * @param {GenerateContext***REMOVED*** generateContext context for generate
	 * @returns {Source | null***REMOVED*** generated code
	 */
	generate(module, { moduleGraph, runtime ***REMOVED***) {
		const bin =
			/** @type {Buffer***REMOVED*** */
			(/** @type {Source***REMOVED*** */ (module.originalSource()).source());

		const initFuncId = t.identifier("");

		// parse it
		const ast = decode(bin, {
			ignoreDataSection: true,
			ignoreCodeSection: true,
			ignoreCustomNameSection: true
		***REMOVED***);

		const moduleContext = moduleContextFromModuleAST(ast.body[0]);

		const importedGlobals = getImportedGlobals(ast);
		const countImportedFunc = getCountImportedFunc(ast);
		const startAtFuncOffset = moduleContext.getStart();
		const nextFuncIndex = getNextFuncIndex(ast, countImportedFunc);
		const nextTypeIndex = getNextTypeIndex(ast);

		const usedDependencyMap = getUsedDependencyMap(
			moduleGraph,
			module,
			this.options.mangleImports
		);
		const externalExports = new Set(
			module.dependencies
				.filter(d => d instanceof WebAssemblyExportImportedDependency)
				.map(d => {
					const wasmDep = /** @type {WebAssemblyExportImportedDependency***REMOVED*** */ (
						d
					);
					return wasmDep.exportName;
				***REMOVED***)
		);

		/** @type {t.Instruction[]***REMOVED*** */
		const additionalInitCode = [];

		const transform = compose(
			rewriteExportNames({
				ast,
				moduleGraph,
				module,
				externalExports,
				runtime
			***REMOVED***),

			removeStartFunc({ ast ***REMOVED***),

			rewriteImportedGlobals({ ast, additionalInitCode ***REMOVED***),

			rewriteImports({
				ast,
				usedDependencyMap
			***REMOVED***),

			addInitFunction({
				ast,
				initFuncId,
				importedGlobals,
				additionalInitCode,
				startAtFuncOffset,
				nextFuncIndex,
				nextTypeIndex
			***REMOVED***)
		);

		const newBin = transform(bin);

		const newBuf = Buffer.from(newBin);

		return new RawSource(newBuf);
	***REMOVED***

	/**
	 * @param {Error***REMOVED*** error the error
	 * @param {NormalModule***REMOVED*** module module for which the code should be generated
	 * @param {GenerateContext***REMOVED*** generateContext context for generate
	 * @returns {Source | null***REMOVED*** generated code
	 */
	generateError(error, module, generateContext) {
		return new RawSource(error.message);
	***REMOVED***
***REMOVED***

module.exports = WebAssemblyGenerator;
