// webpack/lib/wasm-sync/WebAssemblyInInitialChunkError.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
*/

"use strict";

const WebpackError = require("../WebpackError");

/** @typedef {import("../ChunkGraph")***REMOVED*** ChunkGraph */
/** @typedef {import("../Module")***REMOVED*** Module */
/** @typedef {import("../ModuleGraph")***REMOVED*** ModuleGraph */
/** @typedef {import("../RequestShortener")***REMOVED*** RequestShortener */

/**
 * @param {Module***REMOVED*** module module to get chains from
 * @param {ModuleGraph***REMOVED*** moduleGraph the module graph
 * @param {ChunkGraph***REMOVED*** chunkGraph the chunk graph
 * @param {RequestShortener***REMOVED*** requestShortener to make readable identifiers
 * @returns {string[]***REMOVED*** all chains to the module
 */
const getInitialModuleChains = (
	module,
	moduleGraph,
	chunkGraph,
	requestShortener
) => {
	const queue = [
		{ head: module, message: module.readableIdentifier(requestShortener) ***REMOVED***
	];
	/** @type {Set<string>***REMOVED*** */
	const results = new Set();
	/** @type {Set<string>***REMOVED*** */
	const incompleteResults = new Set();
	/** @type {Set<Module>***REMOVED*** */
	const visitedModules = new Set();

	for (const chain of queue) {
		const { head, message ***REMOVED*** = chain;
		let final = true;
		/** @type {Set<Module>***REMOVED*** */
		const alreadyReferencedModules = new Set();
		for (const connection of moduleGraph.getIncomingConnections(head)) {
			const newHead = connection.originModule;
			if (newHead) {
				if (!chunkGraph.getModuleChunks(newHead).some(c => c.canBeInitial()))
					continue;
				final = false;
				if (alreadyReferencedModules.has(newHead)) continue;
				alreadyReferencedModules.add(newHead);
				const moduleName = newHead.readableIdentifier(requestShortener);
				const detail = connection.explanation
					? ` (${connection.explanation***REMOVED***)`
					: "";
				const newMessage = `${moduleName***REMOVED***${detail***REMOVED*** --> ${message***REMOVED***`;
				if (visitedModules.has(newHead)) {
					incompleteResults.add(`... --> ${newMessage***REMOVED***`);
					continue;
				***REMOVED***
				visitedModules.add(newHead);
				queue.push({
					head: newHead,
					message: newMessage
				***REMOVED***);
			***REMOVED*** else {
				final = false;
				const newMessage = connection.explanation
					? `(${connection.explanation***REMOVED***) --> ${message***REMOVED***`
					: message;
				results.add(newMessage);
			***REMOVED***
		***REMOVED***
		if (final) {
			results.add(message);
		***REMOVED***
	***REMOVED***
	for (const result of incompleteResults) {
		results.add(result);
	***REMOVED***
	return Array.from(results);
***REMOVED***;

module.exports = class WebAssemblyInInitialChunkError extends WebpackError {
	/**
	 * @param {Module***REMOVED*** module WASM module
	 * @param {ModuleGraph***REMOVED*** moduleGraph the module graph
	 * @param {ChunkGraph***REMOVED*** chunkGraph the chunk graph
	 * @param {RequestShortener***REMOVED*** requestShortener request shortener
	 */
	constructor(module, moduleGraph, chunkGraph, requestShortener) {
		const moduleChains = getInitialModuleChains(
			module,
			moduleGraph,
			chunkGraph,
			requestShortener
		);
		const message = `WebAssembly module is included in initial chunk.
This is not allowed, because WebAssembly download and compilation must happen asynchronous.
Add an async split point (i. e. import()) somewhere between your entrypoint and the WebAssembly module:
${moduleChains.map(s => `* ${s***REMOVED***`).join("\n")***REMOVED***`;

		super(message);
		this.name = "WebAssemblyInInitialChunkError";
		this.hideStack = true;
		this.module = module;
	***REMOVED***
***REMOVED***;
