// babel-plugin-transform-react-remove-prop-types/src/isStatelessComponent.js
const traversed = Symbol('traversed')

function isJSXElementOrReactCreateElement(path) {
  let visited = false

  path.traverse({
    CallExpression(path2) {
      const callee = path2.get('callee')

      if (
        callee.matchesPattern('React.createElement') ||
        callee.matchesPattern('React.cloneElement') ||
        callee.node.name === 'cloneElement'
      ) {
        visited = true
      ***REMOVED***
    ***REMOVED***,
    JSXElement() {
      visited = true
    ***REMOVED***,
  ***REMOVED***)

  return visited
***REMOVED***

function isReturningJSXElement(path, iteration = 0) {
  // Early exit for ArrowFunctionExpressions, there is no ReturnStatement node.
  if (path.node.init && path.node.init.body && isJSXElementOrReactCreateElement(path)) {
    return true
  ***REMOVED***

  if (iteration > 20) {
    throw new Error('transform-react-remove-prop-type: infinite loop detected.')
  ***REMOVED***

  let visited = false

  path.traverse({
    ReturnStatement(path2) {
      // We have already found what we are looking for.
      if (visited) {
        return
      ***REMOVED***

      const argument = path2.get('argument')

      // Nothing is returned
      if (!argument.node) {
        return
      ***REMOVED***

      if (isJSXElementOrReactCreateElement(path2)) {
        visited = true
        return
      ***REMOVED***

      if (argument.node.type === 'CallExpression') {
        const name = argument.get('callee').node.name
        const binding = path.scope.getBinding(name)

        if (!binding) {
          return
        ***REMOVED***

        // Prevents infinite traverse loop.
        if (binding.path[traversed]) {
          return
        ***REMOVED***

        binding.path[traversed] = true

        if (isReturningJSXElement(binding.path, iteration + 1)) {
          visited = true
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***,
  ***REMOVED***)

  return visited
***REMOVED***

const VALID_POSSIBLE_STATELESS_COMPONENT_TYPES = ['VariableDeclarator', 'FunctionDeclaration']

// Returns `true` if the path represents a function which returns a JSXElement
export default function isStatelessComponent(path) {
  if (VALID_POSSIBLE_STATELESS_COMPONENT_TYPES.indexOf(path.node.type) === -1) {
    return false
  ***REMOVED***

  if (isReturningJSXElement(path)) {
    return true
  ***REMOVED***

  return false
***REMOVED***
