// @babel/helper-create-class-features-plugin/lib/misc.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.extractComputedKeys = extractComputedKeys;
exports.injectInitialization = injectInitialization;
exports.memoiseComputedKey = memoiseComputedKey;
var _core = require("@babel/core");
var _traverse = require("@babel/traverse");
const findBareSupers = _traverse.visitors.environmentVisitor({
  Super(path) {
    const {
      node,
      parentPath
    ***REMOVED*** = path;
    if (parentPath.isCallExpression({
      callee: node
    ***REMOVED***)) {
      this.push(parentPath);
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
const referenceVisitor = {
  "TSTypeAnnotation|TypeAnnotation"(path) {
    path.skip();
  ***REMOVED***,
  ReferencedIdentifier(path, {
    scope
  ***REMOVED***) {
    if (scope.hasOwnBinding(path.node.name)) {
      scope.rename(path.node.name);
      path.skip();
    ***REMOVED***
  ***REMOVED***
***REMOVED***;
function handleClassTDZ(path, state) {
  if (state.classBinding && state.classBinding === path.scope.getBinding(path.node.name)) {
    const classNameTDZError = state.file.addHelper("classNameTDZError");
    const throwNode = _core.types.callExpression(classNameTDZError, [_core.types.stringLiteral(path.node.name)]);
    path.replaceWith(_core.types.sequenceExpression([throwNode, path.node]));
    path.skip();
  ***REMOVED***
***REMOVED***
const classFieldDefinitionEvaluationTDZVisitor = {
  ReferencedIdentifier: handleClassTDZ,
  "TSTypeAnnotation|TypeAnnotation"(path) {
    path.skip();
  ***REMOVED***
***REMOVED***;
function injectInitialization(path, constructor, nodes, renamer, lastReturnsThis) {
  if (!nodes.length) return;
  const isDerived = !!path.node.superClass;
  if (!constructor) {
    const newConstructor = _core.types.classMethod("constructor", _core.types.identifier("constructor"), [], _core.types.blockStatement([]));
    if (isDerived) {
      newConstructor.params = [_core.types.restElement(_core.types.identifier("args"))];
      newConstructor.body.body.push(_core.template.statement.ast`super(...args)`);
    ***REMOVED***
    [constructor] = path.get("body").unshiftContainer("body", newConstructor);
  ***REMOVED***
  if (renamer) {
    renamer(referenceVisitor, {
      scope: constructor.scope
    ***REMOVED***);
  ***REMOVED***
  if (isDerived) {
    const bareSupers = [];
    constructor.traverse(findBareSupers, bareSupers);
    let isFirst = true;
    for (const bareSuper of bareSupers) {
      if (isFirst) {
        isFirst = false;
      ***REMOVED*** else {
        nodes = nodes.map(n => _core.types.cloneNode(n));
      ***REMOVED***
      if (!bareSuper.parentPath.isExpressionStatement()) {
        const allNodes = [bareSuper.node, ...nodes.map(n => _core.types.toExpression(n))];
        if (!lastReturnsThis) allNodes.push(_core.types.thisExpression());
        bareSuper.replaceWith(_core.types.sequenceExpression(allNodes));
      ***REMOVED*** else {
        bareSuper.insertAfter(nodes);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED*** else {
    constructor.get("body").unshiftContainer("body", nodes);
  ***REMOVED***
***REMOVED***
function memoiseComputedKey(keyNode, scope, hint) {
  const isUidReference = _core.types.isIdentifier(keyNode) && scope.hasUid(keyNode.name);
  if (isUidReference) {
    return;
  ***REMOVED***
  const isMemoiseAssignment = _core.types.isAssignmentExpression(keyNode, {
    operator: "="
  ***REMOVED***) && _core.types.isIdentifier(keyNode.left) && scope.hasUid(keyNode.left.name);
  if (isMemoiseAssignment) {
    return _core.types.cloneNode(keyNode);
  ***REMOVED*** else {
    const ident = _core.types.identifier(hint);
    scope.push({
      id: ident,
      kind: "let"
    ***REMOVED***);
    return _core.types.assignmentExpression("=", _core.types.cloneNode(ident), keyNode);
  ***REMOVED***
***REMOVED***
function extractComputedKeys(path, computedPaths, file) {
  const {
    scope
  ***REMOVED*** = path;
  const declarations = [];
  const state = {
    classBinding: path.node.id && scope.getBinding(path.node.id.name),
    file
  ***REMOVED***;
  for (const computedPath of computedPaths) {
    const computedKey = computedPath.get("key");
    if (computedKey.isReferencedIdentifier()) {
      handleClassTDZ(computedKey, state);
    ***REMOVED*** else {
      computedKey.traverse(classFieldDefinitionEvaluationTDZVisitor, state);
    ***REMOVED***
    const computedNode = computedPath.node;
    if (!computedKey.isConstantExpression()) {
      const assignment = memoiseComputedKey(computedKey.node, scope, scope.generateUidBasedOnNode(computedKey.node));
      if (assignment) {
        declarations.push(_core.types.expressionStatement(assignment));
        computedNode.key = _core.types.cloneNode(assignment.left);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  return declarations;
***REMOVED***
