// browserslist/node.js
var feature = require('caniuse-lite/dist/unpacker/feature').default
var region = require('caniuse-lite/dist/unpacker/region').default
var fs = require('fs')
var path = require('path')

var BrowserslistError = require('./error')

var IS_SECTION = /^\s*\[(.+)]\s*$/
var CONFIG_PATTERN = /^browserslist-config-/
var SCOPED_CONFIG__PATTERN = /@[^/]+(?:\/[^/]+)?\/browserslist-config(?:-|$|\/)/
var FORMAT =
  'Browserslist config should be a string or an array ' +
  'of strings with browser queries'

var dataTimeChecked = false
var statCache = {***REMOVED***
var configPathCache = {***REMOVED***
var parseConfigCache = {***REMOVED***

function checkExtend(name) {
  var use = ' Use `dangerousExtend` option to disable.'
  if (!CONFIG_PATTERN.test(name) && !SCOPED_CONFIG__PATTERN.test(name)) {
    throw new BrowserslistError(
      'Browserslist config needs `browserslist-config-` prefix. ' + use
    )
  ***REMOVED***
  if (name.replace(/^@[^/]+\//, '').indexOf('.') !== -1) {
    throw new BrowserslistError(
      '`.` not allowed in Browserslist config name. ' + use
    )
  ***REMOVED***
  if (name.indexOf('node_modules') !== -1) {
    throw new BrowserslistError(
      '`node_modules` not allowed in Browserslist config.' + use
    )
  ***REMOVED***
***REMOVED***

function isFile(file) {
  return fs.existsSync(file) && fs.statSync(file).isFile()
***REMOVED***
function isDirectory(dir) {
  return fs.existsSync(dir) && fs.statSync(dir).isDirectory()
***REMOVED***

function eachParent(file, callback, cache) {
  var loc = path.resolve(file)
  var pathsForCacheResult = []
  var result
  do {
    if (!pathInRoot(loc)) {
      break
    ***REMOVED***
    if (cache && loc in cache) {
      result = cache[loc]
      break
    ***REMOVED***
    pathsForCacheResult.push(loc)

    if (!isDirectory(loc)) {
      continue
    ***REMOVED***

    var locResult = callback(loc)
    if (typeof locResult !== 'undefined') {
      result = locResult
      break
    ***REMOVED***
  ***REMOVED*** while (loc !== (loc = path.dirname(loc)))

  if (cache && !process.env.BROWSERSLIST_DISABLE_CACHE) {
    pathsForCacheResult.forEach(function (cachePath) {
      cache[cachePath] = result
    ***REMOVED***)
  ***REMOVED***
  return result
***REMOVED***

function pathInRoot(p) {
  if (!process.env.BROWSERSLIST_ROOT_PATH) return true
  var rootPath = path.resolve(process.env.BROWSERSLIST_ROOT_PATH)
  if (path.relative(rootPath, p).substring(0, 2) === '..') {
    return false
  ***REMOVED***
  return true
***REMOVED***

function check(section) {
  if (Array.isArray(section)) {
    for (var i = 0; i < section.length; i++) {
      if (typeof section[i] !== 'string') {
        throw new BrowserslistError(FORMAT)
      ***REMOVED***
    ***REMOVED***
  ***REMOVED*** else if (typeof section !== 'string') {
    throw new BrowserslistError(FORMAT)
  ***REMOVED***
***REMOVED***

function pickEnv(config, opts) {
  if (typeof config !== 'object') return config

  var name
  if (typeof opts.env === 'string') {
    name = opts.env
  ***REMOVED*** else if (process.env.BROWSERSLIST_ENV) {
    name = process.env.BROWSERSLIST_ENV
  ***REMOVED*** else if (process.env.NODE_ENV) {
    name = process.env.NODE_ENV
  ***REMOVED*** else {
    name = 'production'
  ***REMOVED***

  if (opts.throwOnMissing) {
    if (name && name !== 'defaults' && !config[name]) {
      throw new BrowserslistError(
        'Missing config for Browserslist environment `' + name + '`'
      )
    ***REMOVED***
  ***REMOVED***

  return config[name] || config.defaults
***REMOVED***

function parsePackage(file) {
  var text = fs
    .readFileSync(file)
    .toString()
    .replace(/^\uFEFF/m, '')
  var list
  if (text.indexOf('"browserslist"') >= 0) {
    list = JSON.parse(text).browserslist
  ***REMOVED*** else if (text.indexOf('"browserlist"') >= 0) {
    var config = JSON.parse(text)
    if (config.browserlist && !config.browserslist) {
      throw new BrowserslistError(
        '`browserlist` key instead of `browserslist` in ' + file
      )
    ***REMOVED***
  ***REMOVED***
  if (Array.isArray(list) || typeof list === 'string') {
    list = { defaults: list ***REMOVED***
  ***REMOVED***
  for (var i in list) {
    check(list[i])
  ***REMOVED***

  return list
***REMOVED***

function parsePackageOrReadConfig(file) {
  if (file in parseConfigCache) {
    return parseConfigCache[file]
  ***REMOVED***

  var isPackage = path.basename(file) === 'package.json'
  var result = isPackage ? parsePackage(file) : module.exports.readConfig(file)

  if (!process.env.BROWSERSLIST_DISABLE_CACHE) {
    parseConfigCache[file] = result
  ***REMOVED***
  return result
***REMOVED***

function latestReleaseTime(agents) {
  var latest = 0
  for (var name in agents) {
    var dates = agents[name].releaseDate || {***REMOVED***
    for (var key in dates) {
      if (latest < dates[key]) {
        latest = dates[key]
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  return latest * 1000
***REMOVED***

function getMonthsPassed(date) {
  var now = new Date()
  var past = new Date(date)

  var years = now.getFullYear() - past.getFullYear()
  var months = now.getMonth() - past.getMonth()

  return years * 12 + months
***REMOVED***

function normalizeStats(data, stats) {
  if (!data) {
    data = {***REMOVED***
  ***REMOVED***
  if (stats && 'dataByBrowser' in stats) {
    stats = stats.dataByBrowser
  ***REMOVED***

  if (typeof stats !== 'object') return undefined

  var normalized = {***REMOVED***
  for (var i in stats) {
    var versions = Object.keys(stats[i])
    if (versions.length === 1 && data[i] && data[i].versions.length === 1) {
      var normal = data[i].versions[0]
      normalized[i] = {***REMOVED***
      normalized[i][normal] = stats[i][versions[0]]
    ***REMOVED*** else {
      normalized[i] = stats[i]
    ***REMOVED***
  ***REMOVED***

  return normalized
***REMOVED***

function normalizeUsageData(usageData, data) {
  for (var browser in usageData) {
    var browserUsage = usageData[browser]
    // https://github.com/browserslist/browserslist/issues/431#issuecomment-565230615
    // caniuse-db returns { 0: "percentage" ***REMOVED*** for `and_*` regional stats
    if ('0' in browserUsage) {
      var versions = data[browser].versions
      browserUsage[versions[versions.length - 1]] = browserUsage[0]
      delete browserUsage[0]
    ***REMOVED***
  ***REMOVED***
***REMOVED***

module.exports = {
  loadQueries: function loadQueries(ctx, name) {
    if (!ctx.dangerousExtend && !process.env.BROWSERSLIST_DANGEROUS_EXTEND) {
      checkExtend(name)
    ***REMOVED***
    var queries = require(require.resolve(name, { paths: ['.', ctx.path] ***REMOVED***))
    if (typeof queries === 'object' && queries !== null && queries.__esModule) {
      queries = queries.default
    ***REMOVED***
    if (queries) {
      if (Array.isArray(queries)) {
        return queries
      ***REMOVED*** else if (typeof queries === 'object') {
        if (!queries.defaults) queries.defaults = []
        return pickEnv(queries, ctx, name)
      ***REMOVED***
    ***REMOVED***
    throw new BrowserslistError(
      '`' +
        name +
        '` config exports not an array of queries' +
        ' or an object of envs'
    )
  ***REMOVED***,

  loadStat: function loadStat(ctx, name, data) {
    if (!ctx.dangerousExtend && !process.env.BROWSERSLIST_DANGEROUS_EXTEND) {
      checkExtend(name)
    ***REMOVED***
    var stats = require(require.resolve(
      path.join(name, 'browserslist-stats.json'),
      { paths: ['.'] ***REMOVED***
    ))
    return normalizeStats(data, stats)
  ***REMOVED***,

  getStat: function getStat(opts, data) {
    var stats
    if (opts.stats) {
      stats = opts.stats
    ***REMOVED*** else if (process.env.BROWSERSLIST_STATS) {
      stats = process.env.BROWSERSLIST_STATS
    ***REMOVED*** else if (opts.path && path.resolve && fs.existsSync) {
      stats = eachParent(
        opts.path,
        function (dir) {
          var file = path.join(dir, 'browserslist-stats.json')
          return isFile(file) ? file : undefined
        ***REMOVED***,
        statCache
      )
    ***REMOVED***
    if (typeof stats === 'string') {
      try {
        stats = JSON.parse(fs.readFileSync(stats))
      ***REMOVED*** catch (e) {
        throw new BrowserslistError("Can't read " + stats)
      ***REMOVED***
    ***REMOVED***
    return normalizeStats(data, stats)
  ***REMOVED***,

  loadConfig: function loadConfig(opts) {
    if (process.env.BROWSERSLIST) {
      return process.env.BROWSERSLIST
    ***REMOVED*** else if (opts.config || process.env.BROWSERSLIST_CONFIG) {
      var file = opts.config || process.env.BROWSERSLIST_CONFIG
      return pickEnv(parsePackageOrReadConfig(file), opts)
    ***REMOVED*** else if (opts.path) {
      return pickEnv(module.exports.findConfig(opts.path), opts)
    ***REMOVED*** else {
      return undefined
    ***REMOVED***
  ***REMOVED***,

  loadCountry: function loadCountry(usage, country, data) {
    var code = country.replace(/[^\w-]/g, '')
    if (!usage[code]) {
      var compressed
      try {
        compressed = require('caniuse-lite/data/regions/' + code + '.js')
      ***REMOVED*** catch (e) {
        throw new BrowserslistError('Unknown region name `' + code + '`.')
      ***REMOVED***
      var usageData = region(compressed)
      normalizeUsageData(usageData, data)
      usage[country] = {***REMOVED***
      for (var i in usageData) {
        for (var j in usageData[i]) {
          usage[country][i + ' ' + j] = usageData[i][j]
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***,

  loadFeature: function loadFeature(features, name) {
    name = name.replace(/[^\w-]/g, '')
    if (features[name]) return
    var compressed
    try {
      compressed = require('caniuse-lite/data/features/' + name + '.js')
    ***REMOVED*** catch (e) {
      throw new BrowserslistError('Unknown feature name `' + name + '`.')
    ***REMOVED***
    var stats = feature(compressed).stats
    features[name] = {***REMOVED***
    for (var i in stats) {
      features[name][i] = {***REMOVED***
      for (var j in stats[i]) {
        features[name][i][j] = stats[i][j]
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***,

  parseConfig: function parseConfig(string) {
    var result = { defaults: [] ***REMOVED***
    var sections = ['defaults']

    string
      .toString()
      .replace(/#[^\n]*/g, '')
      .split(/\n|,/)
      .map(function (line) {
        return line.trim()
      ***REMOVED***)
      .filter(function (line) {
        return line !== ''
      ***REMOVED***)
      .forEach(function (line) {
        if (IS_SECTION.test(line)) {
          sections = line.match(IS_SECTION)[1].trim().split(' ')
          sections.forEach(function (section) {
            if (result[section]) {
              throw new BrowserslistError(
                'Duplicate section ' + section + ' in Browserslist config'
              )
            ***REMOVED***
            result[section] = []
          ***REMOVED***)
        ***REMOVED*** else {
          sections.forEach(function (section) {
            result[section].push(line)
          ***REMOVED***)
        ***REMOVED***
      ***REMOVED***)

    return result
  ***REMOVED***,

  readConfig: function readConfig(file) {
    if (!isFile(file)) {
      throw new BrowserslistError("Can't read " + file + ' config')
    ***REMOVED***

    return module.exports.parseConfig(fs.readFileSync(file))
  ***REMOVED***,

  findConfigFile: function findConfigFile(from) {
    return eachParent(
      from,
      function (dir) {
        var config = path.join(dir, 'browserslist')
        var pkg = path.join(dir, 'package.json')
        var rc = path.join(dir, '.browserslistrc')

        var pkgBrowserslist
        if (isFile(pkg)) {
          try {
            pkgBrowserslist = parsePackage(pkg)
          ***REMOVED*** catch (e) {
            if (e.name === 'BrowserslistError') throw e
            console.warn(
              '[Browserslist] Could not parse ' + pkg + '. Ignoring it.'
            )
          ***REMOVED***
        ***REMOVED***

        if (isFile(config) && pkgBrowserslist) {
          throw new BrowserslistError(
            dir + ' contains both browserslist and package.json with browsers'
          )
        ***REMOVED*** else if (isFile(rc) && pkgBrowserslist) {
          throw new BrowserslistError(
            dir +
              ' contains both .browserslistrc and package.json with browsers'
          )
        ***REMOVED*** else if (isFile(config) && isFile(rc)) {
          throw new BrowserslistError(
            dir + ' contains both .browserslistrc and browserslist'
          )
        ***REMOVED*** else if (isFile(config)) {
          return config
        ***REMOVED*** else if (isFile(rc)) {
          return rc
        ***REMOVED*** else if (pkgBrowserslist) {
          return pkg
        ***REMOVED***
      ***REMOVED***,
      configPathCache
    )
  ***REMOVED***,

  findConfig: function findConfig(from) {
    var configFile = this.findConfigFile(from)

    return configFile ? parsePackageOrReadConfig(configFile) : undefined
  ***REMOVED***,

  clearCaches: function clearCaches() {
    dataTimeChecked = false
    statCache = {***REMOVED***
    configPathCache = {***REMOVED***
    parseConfigCache = {***REMOVED***

    this.cache = {***REMOVED***
  ***REMOVED***,

  oldDataWarning: function oldDataWarning(agentsObj) {
    if (dataTimeChecked) return
    dataTimeChecked = true
    if (process.env.BROWSERSLIST_IGNORE_OLD_DATA) return

    var latest = latestReleaseTime(agentsObj)
    var monthsPassed = getMonthsPassed(latest)

    if (latest !== 0 && monthsPassed >= 6) {
      var months = monthsPassed + ' ' + (monthsPassed > 1 ? 'months' : 'month')
      console.warn(
        'Browserslist: browsers data (caniuse-lite) is ' +
          months +
          ' old. Please run:\n' +
          '  npx update-browserslist-db@latest\n' +
          '  Why you should do it regularly: ' +
          'https://github.com/browserslist/update-db#readme'
      )
    ***REMOVED***
  ***REMOVED***,

  currentNode: function currentNode() {
    return 'node ' + process.versions.node
  ***REMOVED***,

  env: process.env
***REMOVED***
