// throat/index.js
'use strict';

function throatInternal(size) {
  var queue = new Queue();
  var s = size | 0;

  function run(fn, self, args) {
    if ((s | 0) !== 0) {
      s = (s | 0) - 1;
      return new Promise(function (resolve) {
        resolve(fn.apply(self, args));
      ***REMOVED***).then(onFulfill, onReject);
    ***REMOVED***
    return new Promise(function (resolve) {
      queue.push(new Delayed(resolve, fn, self, args));
    ***REMOVED***).then(runDelayed);
  ***REMOVED***
  function runDelayed(d) {
    try {
      return Promise.resolve(d.fn.apply(d.self, d.args)).then(
        onFulfill,
        onReject
      );
    ***REMOVED*** catch (ex) {
      onReject(ex);
    ***REMOVED***
  ***REMOVED***
  function onFulfill(result) {
    release();
    return result;
  ***REMOVED***
  function onReject(error) {
    release();
    throw error;
  ***REMOVED***
  function release() {
    var next = queue.shift();
    if (next) {
      next.resolve(next);
    ***REMOVED*** else {
      s = (s | 0) + 1;
    ***REMOVED***
  ***REMOVED***

  return run;
***REMOVED***

function earlyBound(size, fn) {
  const run = throatInternal(size | 0);
  return function () {
    var args = new Array(arguments.length);
    for (var i = 0; i < arguments.length; i++) {
      args[i] = arguments[i];
    ***REMOVED***
    return run(fn, this, args);
  ***REMOVED***;
***REMOVED***
function lateBound(size) {
  const run = throatInternal(size | 0);
  return function (fn) {
    if (typeof fn !== 'function') {
      throw new TypeError(
        'Expected throat fn to be a function but got ' + typeof fn
      );
    ***REMOVED***
    var args = new Array(arguments.length - 1);
    for (var i = 1; i < arguments.length; i++) {
      args[i - 1] = arguments[i];
    ***REMOVED***
    return run(fn, this, args);
  ***REMOVED***;
***REMOVED***
module.exports = function throat(size, fn) {
  if (typeof size === 'function') {
    var temp = fn;
    fn = size;
    size = temp;
  ***REMOVED***
  if (typeof size !== 'number') {
    throw new TypeError(
      'Expected throat size to be a number but got ' + typeof size
    );
  ***REMOVED***
  if (fn !== undefined && typeof fn !== 'function') {
    throw new TypeError(
      'Expected throat fn to be a function but got ' + typeof fn
    );
  ***REMOVED***
  if (typeof fn === 'function') {
    return earlyBound(size | 0, fn);
  ***REMOVED*** else {
    return lateBound(size | 0);
  ***REMOVED***
***REMOVED***;

module.exports.default = module.exports;

function Delayed(resolve, fn, self, args) {
  this.resolve = resolve;
  this.fn = fn;
  this.self = self || null;
  this.args = args;
***REMOVED***

var blockSize = 64;
function Queue() {
  this._s1 = [];
  this._s2 = [];
  this._shiftBlock = this._pushBlock = new Array(blockSize);
  this._pushIndex = 0;
  this._shiftIndex = 0;
***REMOVED***

Queue.prototype.push = function (value) {
  if (this._pushIndex === blockSize) {
    this._pushIndex = 0;
    this._s1[this._s1.length] = this._pushBlock = new Array(blockSize);
  ***REMOVED***
  this._pushBlock[this._pushIndex++] = value;
***REMOVED***;

Queue.prototype.shift = function () {
  if (this._shiftIndex === blockSize) {
    var s2 = this._s2;
    if (s2.length === 0) {
      var s1 = this._s1;
      if (s1.length === 0) {
        return undefined;
      ***REMOVED***
      this._s1 = s2;
      s2 = this._s2 = s1.reverse();
    ***REMOVED***
    this._shiftIndex = 0;
    this._shiftBlock = s2.pop();
  ***REMOVED***
  if (
    this._pushBlock === this._shiftBlock &&
    this._pushIndex === this._shiftIndex
  ) {
    return undefined;
  ***REMOVED***
  var result = this._shiftBlock[this._shiftIndex];
  this._shiftBlock[this._shiftIndex++] = null;
  return result;
***REMOVED***;
