// memfs/lib/node.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] ***REMOVED*** instanceof Array && function (d, b) { d.__proto__ = b; ***REMOVED***) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; ***REMOVED***;
        return extendStatics(d, b);
    ***REMOVED***;
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.File = exports.Link = exports.Node = exports.SEP = void 0;
var process_1 = require("./process");
var buffer_1 = require("./internal/buffer");
var constants_1 = require("./constants");
var events_1 = require("events");
var Stats_1 = require("./Stats");
var S_IFMT = constants_1.constants.S_IFMT, S_IFDIR = constants_1.constants.S_IFDIR, S_IFREG = constants_1.constants.S_IFREG, S_IFLNK = constants_1.constants.S_IFLNK, O_APPEND = constants_1.constants.O_APPEND;
var getuid = function () { var _a, _b; return (_b = (_a = process_1.default.getuid) === null || _a === void 0 ? void 0 : _a.call(process_1.default)) !== null && _b !== void 0 ? _b : 0; ***REMOVED***;
var getgid = function () { var _a, _b; return (_b = (_a = process_1.default.getgid) === null || _a === void 0 ? void 0 : _a.call(process_1.default)) !== null && _b !== void 0 ? _b : 0; ***REMOVED***;
exports.SEP = '/';
/**
 * Node in a file system (like i-node, v-node).
 */
var Node = /** @class */ (function (_super) {
    __extends(Node, _super);
    function Node(ino, perm) {
        if (perm === void 0) { perm = 438; ***REMOVED***
        var _this = _super.call(this) || this;
        // User ID and group ID.
        _this._uid = getuid();
        _this._gid = getgid();
        _this._atime = new Date();
        _this._mtime = new Date();
        _this._ctime = new Date();
        _this._perm = 438; // Permissions `chmod`, `fchmod`
        _this.mode = S_IFREG; // S_IFDIR, S_IFREG, etc.. (file by default?)
        // Number of hard links pointing at this Node.
        _this._nlink = 1;
        _this._perm = perm;
        _this.mode |= perm;
        _this.ino = ino;
        return _this;
    ***REMOVED***
    Object.defineProperty(Node.prototype, "ctime", {
        get: function () {
            return this._ctime;
        ***REMOVED***,
        set: function (ctime) {
            this._ctime = ctime;
        ***REMOVED***,
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(Node.prototype, "uid", {
        get: function () {
            return this._uid;
        ***REMOVED***,
        set: function (uid) {
            this._uid = uid;
            this.ctime = new Date();
        ***REMOVED***,
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(Node.prototype, "gid", {
        get: function () {
            return this._gid;
        ***REMOVED***,
        set: function (gid) {
            this._gid = gid;
            this.ctime = new Date();
        ***REMOVED***,
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(Node.prototype, "atime", {
        get: function () {
            return this._atime;
        ***REMOVED***,
        set: function (atime) {
            this._atime = atime;
            this.ctime = new Date();
        ***REMOVED***,
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(Node.prototype, "mtime", {
        get: function () {
            return this._mtime;
        ***REMOVED***,
        set: function (mtime) {
            this._mtime = mtime;
            this.ctime = new Date();
        ***REMOVED***,
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(Node.prototype, "perm", {
        get: function () {
            return this._perm;
        ***REMOVED***,
        set: function (perm) {
            this._perm = perm;
            this.ctime = new Date();
        ***REMOVED***,
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(Node.prototype, "nlink", {
        get: function () {
            return this._nlink;
        ***REMOVED***,
        set: function (nlink) {
            this._nlink = nlink;
            this.ctime = new Date();
        ***REMOVED***,
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Node.prototype.getString = function (encoding) {
        if (encoding === void 0) { encoding = 'utf8'; ***REMOVED***
        this.atime = new Date();
        return this.getBuffer().toString(encoding);
    ***REMOVED***;
    Node.prototype.setString = function (str) {
        // this.setBuffer(bufferFrom(str, 'utf8'));
        this.buf = (0, buffer_1.bufferFrom)(str, 'utf8');
        this.touch();
    ***REMOVED***;
    Node.prototype.getBuffer = function () {
        this.atime = new Date();
        if (!this.buf)
            this.setBuffer((0, buffer_1.bufferAllocUnsafe)(0));
        return (0, buffer_1.bufferFrom)(this.buf); // Return a copy.
    ***REMOVED***;
    Node.prototype.setBuffer = function (buf) {
        this.buf = (0, buffer_1.bufferFrom)(buf); // Creates a copy of data.
        this.touch();
    ***REMOVED***;
    Node.prototype.getSize = function () {
        return this.buf ? this.buf.length : 0;
    ***REMOVED***;
    Node.prototype.setModeProperty = function (property) {
        this.mode = (this.mode & ~S_IFMT) | property;
    ***REMOVED***;
    Node.prototype.setIsFile = function () {
        this.setModeProperty(S_IFREG);
    ***REMOVED***;
    Node.prototype.setIsDirectory = function () {
        this.setModeProperty(S_IFDIR);
    ***REMOVED***;
    Node.prototype.setIsSymlink = function () {
        this.setModeProperty(S_IFLNK);
    ***REMOVED***;
    Node.prototype.isFile = function () {
        return (this.mode & S_IFMT) === S_IFREG;
    ***REMOVED***;
    Node.prototype.isDirectory = function () {
        return (this.mode & S_IFMT) === S_IFDIR;
    ***REMOVED***;
    Node.prototype.isSymlink = function () {
        // return !!this.symlink;
        return (this.mode & S_IFMT) === S_IFLNK;
    ***REMOVED***;
    Node.prototype.makeSymlink = function (steps) {
        this.symlink = steps;
        this.setIsSymlink();
    ***REMOVED***;
    Node.prototype.write = function (buf, off, len, pos) {
        if (off === void 0) { off = 0; ***REMOVED***
        if (len === void 0) { len = buf.length; ***REMOVED***
        if (pos === void 0) { pos = 0; ***REMOVED***
        if (!this.buf)
            this.buf = (0, buffer_1.bufferAllocUnsafe)(0);
        if (pos + len > this.buf.length) {
            var newBuf = (0, buffer_1.bufferAllocUnsafe)(pos + len);
            this.buf.copy(newBuf, 0, 0, this.buf.length);
            this.buf = newBuf;
        ***REMOVED***
        buf.copy(this.buf, pos, off, off + len);
        this.touch();
        return len;
    ***REMOVED***;
    // Returns the number of bytes read.
    Node.prototype.read = function (buf, off, len, pos) {
        if (off === void 0) { off = 0; ***REMOVED***
        if (len === void 0) { len = buf.byteLength; ***REMOVED***
        if (pos === void 0) { pos = 0; ***REMOVED***
        this.atime = new Date();
        if (!this.buf)
            this.buf = (0, buffer_1.bufferAllocUnsafe)(0);
        var actualLen = len;
        if (actualLen > buf.byteLength) {
            actualLen = buf.byteLength;
        ***REMOVED***
        if (actualLen + pos > this.buf.length) {
            actualLen = this.buf.length - pos;
        ***REMOVED***
        this.buf.copy(buf, off, pos, pos + actualLen);
        return actualLen;
    ***REMOVED***;
    Node.prototype.truncate = function (len) {
        if (len === void 0) { len = 0; ***REMOVED***
        if (!len)
            this.buf = (0, buffer_1.bufferAllocUnsafe)(0);
        else {
            if (!this.buf)
                this.buf = (0, buffer_1.bufferAllocUnsafe)(0);
            if (len <= this.buf.length) {
                this.buf = this.buf.slice(0, len);
            ***REMOVED***
            else {
                var buf = (0, buffer_1.bufferAllocUnsafe)(len);
                this.buf.copy(buf);
                buf.fill(0, this.buf.length);
                this.buf = buf;
            ***REMOVED***
        ***REMOVED***
        this.touch();
    ***REMOVED***;
    Node.prototype.chmod = function (perm) {
        this.perm = perm;
        this.mode = (this.mode & ~511) | perm;
        this.touch();
    ***REMOVED***;
    Node.prototype.chown = function (uid, gid) {
        this.uid = uid;
        this.gid = gid;
        this.touch();
    ***REMOVED***;
    Node.prototype.touch = function () {
        this.mtime = new Date();
        this.emit('change', this);
    ***REMOVED***;
    Node.prototype.canRead = function (uid, gid) {
        if (uid === void 0) { uid = getuid(); ***REMOVED***
        if (gid === void 0) { gid = getgid(); ***REMOVED***
        if (this.perm & 4 /* S.IROTH */) {
            return true;
        ***REMOVED***
        if (gid === this.gid) {
            if (this.perm & 32 /* S.IRGRP */) {
                return true;
            ***REMOVED***
        ***REMOVED***
        if (uid === this.uid) {
            if (this.perm & 256 /* S.IRUSR */) {
                return true;
            ***REMOVED***
        ***REMOVED***
        return false;
    ***REMOVED***;
    Node.prototype.canWrite = function (uid, gid) {
        if (uid === void 0) { uid = getuid(); ***REMOVED***
        if (gid === void 0) { gid = getgid(); ***REMOVED***
        if (this.perm & 2 /* S.IWOTH */) {
            return true;
        ***REMOVED***
        if (gid === this.gid) {
            if (this.perm & 16 /* S.IWGRP */) {
                return true;
            ***REMOVED***
        ***REMOVED***
        if (uid === this.uid) {
            if (this.perm & 128 /* S.IWUSR */) {
                return true;
            ***REMOVED***
        ***REMOVED***
        return false;
    ***REMOVED***;
    Node.prototype.del = function () {
        this.emit('delete', this);
    ***REMOVED***;
    Node.prototype.toJSON = function () {
        return {
            ino: this.ino,
            uid: this.uid,
            gid: this.gid,
            atime: this.atime.getTime(),
            mtime: this.mtime.getTime(),
            ctime: this.ctime.getTime(),
            perm: this.perm,
            mode: this.mode,
            nlink: this.nlink,
            symlink: this.symlink,
            data: this.getString(),
        ***REMOVED***;
    ***REMOVED***;
    return Node;
***REMOVED***(events_1.EventEmitter));
exports.Node = Node;
/**
 * Represents a hard link that points to an i-node `node`.
 */
var Link = /** @class */ (function (_super) {
    __extends(Link, _super);
    function Link(vol, parent, name) {
        var _this = _super.call(this) || this;
        _this.children = {***REMOVED***;
        // Path to this node as Array: ['usr', 'bin', 'node'].
        _this._steps = [];
        // "i-node" number of the node.
        _this.ino = 0;
        // Number of children.
        _this.length = 0;
        _this.vol = vol;
        _this.parent = parent;
        _this.name = name;
        _this.syncSteps();
        return _this;
    ***REMOVED***
    Object.defineProperty(Link.prototype, "steps", {
        get: function () {
            return this._steps;
        ***REMOVED***,
        // Recursively sync children steps, e.g. in case of dir rename
        set: function (val) {
            this._steps = val;
            for (var _i = 0, _a = Object.entries(this.children); _i < _a.length; _i++) {
                var _b = _a[_i], child = _b[0], link = _b[1];
                if (child === '.' || child === '..') {
                    continue;
                ***REMOVED***
                link === null || link === void 0 ? void 0 : link.syncSteps();
            ***REMOVED***
        ***REMOVED***,
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Link.prototype.setNode = function (node) {
        this.node = node;
        this.ino = node.ino;
    ***REMOVED***;
    Link.prototype.getNode = function () {
        return this.node;
    ***REMOVED***;
    Link.prototype.createChild = function (name, node) {
        if (node === void 0) { node = this.vol.createNode(); ***REMOVED***
        var link = new Link(this.vol, this, name);
        link.setNode(node);
        if (node.isDirectory()) {
            link.children['.'] = link;
            link.getNode().nlink++;
        ***REMOVED***
        this.setChild(name, link);
        return link;
    ***REMOVED***;
    Link.prototype.setChild = function (name, link) {
        if (link === void 0) { link = new Link(this.vol, this, name); ***REMOVED***
        this.children[name] = link;
        link.parent = this;
        this.length++;
        var node = link.getNode();
        if (node.isDirectory()) {
            link.children['..'] = this;
            this.getNode().nlink++;
        ***REMOVED***
        this.getNode().mtime = new Date();
        this.emit('child:add', link, this);
        return link;
    ***REMOVED***;
    Link.prototype.deleteChild = function (link) {
        var node = link.getNode();
        if (node.isDirectory()) {
            delete link.children['..'];
            this.getNode().nlink--;
        ***REMOVED***
        delete this.children[link.getName()];
        this.length--;
        this.getNode().mtime = new Date();
        this.emit('child:delete', link, this);
    ***REMOVED***;
    Link.prototype.getChild = function (name) {
        this.getNode().mtime = new Date();
        if (Object.hasOwnProperty.call(this.children, name)) {
            return this.children[name];
        ***REMOVED***
    ***REMOVED***;
    Link.prototype.getPath = function () {
        return this.steps.join(exports.SEP);
    ***REMOVED***;
    Link.prototype.getName = function () {
        return this.steps[this.steps.length - 1];
    ***REMOVED***;
    // del() {
    //     const parent = this.parent;
    //     if(parent) {
    //         parent.deleteChild(link);
    //     ***REMOVED***
    //     this.parent = null;
    //     this.vol = null;
    // ***REMOVED***
    /**
     * Walk the tree path and return the `Link` at that location, if any.
     * @param steps {string[]***REMOVED*** Desired location.
     * @param stop {number***REMOVED*** Max steps to go into.
     * @param i {number***REMOVED*** Current step in the `steps` array.
     *
     * @return {Link|null***REMOVED***
     */
    Link.prototype.walk = function (steps, stop, i) {
        if (stop === void 0) { stop = steps.length; ***REMOVED***
        if (i === void 0) { i = 0; ***REMOVED***
        if (i >= steps.length)
            return this;
        if (i >= stop)
            return this;
        var step = steps[i];
        var link = this.getChild(step);
        if (!link)
            return null;
        return link.walk(steps, stop, i + 1);
    ***REMOVED***;
    Link.prototype.toJSON = function () {
        return {
            steps: this.steps,
            ino: this.ino,
            children: Object.keys(this.children),
        ***REMOVED***;
    ***REMOVED***;
    Link.prototype.syncSteps = function () {
        this.steps = this.parent ? this.parent.steps.concat([this.name]) : [this.name];
    ***REMOVED***;
    return Link;
***REMOVED***(events_1.EventEmitter));
exports.Link = Link;
/**
 * Represents an open file (file descriptor) that points to a `Link` (Hard-link) and a `Node`.
 */
var File = /** @class */ (function () {
    /**
     * Open a Link-Node pair. `node` is provided separately as that might be a different node
     * rather the one `link` points to, because it might be a symlink.
     * @param link
     * @param node
     * @param flags
     * @param fd
     */
    function File(link, node, flags, fd) {
        /**
         * A cursor/offset position in a file, where data will be written on write.
         * User can "seek" this position.
         */
        this.position = 0;
        this.link = link;
        this.node = node;
        this.flags = flags;
        this.fd = fd;
    ***REMOVED***
    File.prototype.getString = function (encoding) {
        if (encoding === void 0) { encoding = 'utf8'; ***REMOVED***
        return this.node.getString();
    ***REMOVED***;
    File.prototype.setString = function (str) {
        this.node.setString(str);
    ***REMOVED***;
    File.prototype.getBuffer = function () {
        return this.node.getBuffer();
    ***REMOVED***;
    File.prototype.setBuffer = function (buf) {
        this.node.setBuffer(buf);
    ***REMOVED***;
    File.prototype.getSize = function () {
        return this.node.getSize();
    ***REMOVED***;
    File.prototype.truncate = function (len) {
        this.node.truncate(len);
    ***REMOVED***;
    File.prototype.seekTo = function (position) {
        this.position = position;
    ***REMOVED***;
    File.prototype.stats = function () {
        return Stats_1.default.build(this.node);
    ***REMOVED***;
    File.prototype.write = function (buf, offset, length, position) {
        if (offset === void 0) { offset = 0; ***REMOVED***
        if (length === void 0) { length = buf.length; ***REMOVED***
        if (typeof position !== 'number')
            position = this.position;
        if (this.flags & O_APPEND)
            position = this.getSize();
        var bytes = this.node.write(buf, offset, length, position);
        this.position = position + bytes;
        return bytes;
    ***REMOVED***;
    File.prototype.read = function (buf, offset, length, position) {
        if (offset === void 0) { offset = 0; ***REMOVED***
        if (length === void 0) { length = buf.byteLength; ***REMOVED***
        if (typeof position !== 'number')
            position = this.position;
        var bytes = this.node.read(buf, offset, length, position);
        this.position = position + bytes;
        return bytes;
    ***REMOVED***;
    File.prototype.chmod = function (perm) {
        this.node.chmod(perm);
    ***REMOVED***;
    File.prototype.chown = function (uid, gid) {
        this.node.chown(uid, gid);
    ***REMOVED***;
    return File;
***REMOVED***());
exports.File = File;
