// memfs/lib/promises.js
"use strict";
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        ***REMOVED***
    ***REMOVED***
    return to.concat(ar || Array.prototype.slice.call(from));
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.FileHandle = void 0;
function promisify(vol, fn, getResult) {
    if (getResult === void 0) { getResult = function (input) { return input; ***REMOVED***; ***REMOVED***
    return function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        ***REMOVED***
        return new Promise(function (resolve, reject) {
            vol[fn].bind(vol).apply(void 0, __spreadArray(__spreadArray([], args, false), [function (error, result) {
                    if (error)
                        return reject(error);
                    return resolve(getResult(result));
                ***REMOVED***], false));
        ***REMOVED***);
    ***REMOVED***;
***REMOVED***
var FileHandle = /** @class */ (function () {
    function FileHandle(vol, fd) {
        this.vol = vol;
        this.fd = fd;
    ***REMOVED***
    FileHandle.prototype.appendFile = function (data, options) {
        return promisify(this.vol, 'appendFile')(this.fd, data, options);
    ***REMOVED***;
    FileHandle.prototype.chmod = function (mode) {
        return promisify(this.vol, 'fchmod')(this.fd, mode);
    ***REMOVED***;
    FileHandle.prototype.chown = function (uid, gid) {
        return promisify(this.vol, 'fchown')(this.fd, uid, gid);
    ***REMOVED***;
    FileHandle.prototype.close = function () {
        return promisify(this.vol, 'close')(this.fd);
    ***REMOVED***;
    FileHandle.prototype.datasync = function () {
        return promisify(this.vol, 'fdatasync')(this.fd);
    ***REMOVED***;
    FileHandle.prototype.read = function (buffer, offset, length, position) {
        return promisify(this.vol, 'read', function (bytesRead) { return ({ bytesRead: bytesRead, buffer: buffer ***REMOVED***); ***REMOVED***)(this.fd, buffer, offset, length, position);
    ***REMOVED***;
    FileHandle.prototype.readFile = function (options) {
        return promisify(this.vol, 'readFile')(this.fd, options);
    ***REMOVED***;
    FileHandle.prototype.stat = function (options) {
        return promisify(this.vol, 'fstat')(this.fd, options);
    ***REMOVED***;
    FileHandle.prototype.sync = function () {
        return promisify(this.vol, 'fsync')(this.fd);
    ***REMOVED***;
    FileHandle.prototype.truncate = function (len) {
        return promisify(this.vol, 'ftruncate')(this.fd, len);
    ***REMOVED***;
    FileHandle.prototype.utimes = function (atime, mtime) {
        return promisify(this.vol, 'futimes')(this.fd, atime, mtime);
    ***REMOVED***;
    FileHandle.prototype.write = function (buffer, offset, length, position) {
        return promisify(this.vol, 'write', function (bytesWritten) { return ({ bytesWritten: bytesWritten, buffer: buffer ***REMOVED***); ***REMOVED***)(this.fd, buffer, offset, length, position);
    ***REMOVED***;
    FileHandle.prototype.writeFile = function (data, options) {
        return promisify(this.vol, 'writeFile')(this.fd, data, options);
    ***REMOVED***;
    return FileHandle;
***REMOVED***());
exports.FileHandle = FileHandle;
function createPromisesApi(vol) {
    if (typeof Promise === 'undefined')
        return null;
    return {
        FileHandle: FileHandle,
        access: function (path, mode) {
            return promisify(vol, 'access')(path, mode);
        ***REMOVED***,
        appendFile: function (path, data, options) {
            return promisify(vol, 'appendFile')(path instanceof FileHandle ? path.fd : path, data, options);
        ***REMOVED***,
        chmod: function (path, mode) {
            return promisify(vol, 'chmod')(path, mode);
        ***REMOVED***,
        chown: function (path, uid, gid) {
            return promisify(vol, 'chown')(path, uid, gid);
        ***REMOVED***,
        copyFile: function (src, dest, flags) {
            return promisify(vol, 'copyFile')(src, dest, flags);
        ***REMOVED***,
        lchmod: function (path, mode) {
            return promisify(vol, 'lchmod')(path, mode);
        ***REMOVED***,
        lchown: function (path, uid, gid) {
            return promisify(vol, 'lchown')(path, uid, gid);
        ***REMOVED***,
        link: function (existingPath, newPath) {
            return promisify(vol, 'link')(existingPath, newPath);
        ***REMOVED***,
        lstat: function (path, options) {
            return promisify(vol, 'lstat')(path, options);
        ***REMOVED***,
        mkdir: function (path, options) {
            return promisify(vol, 'mkdir')(path, options);
        ***REMOVED***,
        mkdtemp: function (prefix, options) {
            return promisify(vol, 'mkdtemp')(prefix, options);
        ***REMOVED***,
        open: function (path, flags, mode) {
            return promisify(vol, 'open', function (fd) { return new FileHandle(vol, fd); ***REMOVED***)(path, flags, mode);
        ***REMOVED***,
        readdir: function (path, options) {
            return promisify(vol, 'readdir')(path, options);
        ***REMOVED***,
        readFile: function (id, options) {
            return promisify(vol, 'readFile')(id instanceof FileHandle ? id.fd : id, options);
        ***REMOVED***,
        readlink: function (path, options) {
            return promisify(vol, 'readlink')(path, options);
        ***REMOVED***,
        realpath: function (path, options) {
            return promisify(vol, 'realpath')(path, options);
        ***REMOVED***,
        rename: function (oldPath, newPath) {
            return promisify(vol, 'rename')(oldPath, newPath);
        ***REMOVED***,
        rmdir: function (path) {
            return promisify(vol, 'rmdir')(path);
        ***REMOVED***,
        rm: function (path, options) {
            return promisify(vol, 'rm')(path, options);
        ***REMOVED***,
        stat: function (path, options) {
            return promisify(vol, 'stat')(path, options);
        ***REMOVED***,
        symlink: function (target, path, type) {
            return promisify(vol, 'symlink')(target, path, type);
        ***REMOVED***,
        truncate: function (path, len) {
            return promisify(vol, 'truncate')(path, len);
        ***REMOVED***,
        unlink: function (path) {
            return promisify(vol, 'unlink')(path);
        ***REMOVED***,
        utimes: function (path, atime, mtime) {
            return promisify(vol, 'utimes')(path, atime, mtime);
        ***REMOVED***,
        writeFile: function (id, data, options) {
            return promisify(vol, 'writeFile')(id instanceof FileHandle ? id.fd : id, data, options);
        ***REMOVED***,
    ***REMOVED***;
***REMOVED***
exports.default = createPromisesApi;
