// ajv-formats/dist/limit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.formatLimitDefinition = void 0;
const ajv_1 = require("ajv");
const codegen_1 = require("ajv/dist/compile/codegen");
const ops = codegen_1.operators;
const KWDs = {
    formatMaximum: { okStr: "<=", ok: ops.LTE, fail: ops.GT ***REMOVED***,
    formatMinimum: { okStr: ">=", ok: ops.GTE, fail: ops.LT ***REMOVED***,
    formatExclusiveMaximum: { okStr: "<", ok: ops.LT, fail: ops.GTE ***REMOVED***,
    formatExclusiveMinimum: { okStr: ">", ok: ops.GT, fail: ops.LTE ***REMOVED***,
***REMOVED***;
const error = {
    message: ({ keyword, schemaCode ***REMOVED***) => codegen_1.str `should be ${KWDs[keyword].okStr***REMOVED*** ${schemaCode***REMOVED***`,
    params: ({ keyword, schemaCode ***REMOVED***) => codegen_1._ `{comparison: ${KWDs[keyword].okStr***REMOVED***, limit: ${schemaCode***REMOVED******REMOVED***`,
***REMOVED***;
exports.formatLimitDefinition = {
    keyword: Object.keys(KWDs),
    type: "string",
    schemaType: "string",
    $data: true,
    error,
    code(cxt) {
        const { gen, data, schemaCode, keyword, it ***REMOVED*** = cxt;
        const { opts, self ***REMOVED*** = it;
        if (!opts.validateFormats)
            return;
        const fCxt = new ajv_1.KeywordCxt(it, self.RULES.all.format.definition, "format");
        if (fCxt.$data)
            validate$DataFormat();
        else
            validateFormat();
        function validate$DataFormat() {
            const fmts = gen.scopeValue("formats", {
                ref: self.formats,
                code: opts.code.formats,
            ***REMOVED***);
            const fmt = gen.const("fmt", codegen_1._ `${fmts***REMOVED***[${fCxt.schemaCode***REMOVED***]`);
            cxt.fail$data(codegen_1.or(codegen_1._ `typeof ${fmt***REMOVED*** != "object"`, codegen_1._ `${fmt***REMOVED*** instanceof RegExp`, codegen_1._ `typeof ${fmt***REMOVED***.compare != "function"`, compareCode(fmt)));
        ***REMOVED***
        function validateFormat() {
            const format = fCxt.schema;
            const fmtDef = self.formats[format];
            if (!fmtDef || fmtDef === true)
                return;
            if (typeof fmtDef != "object" ||
                fmtDef instanceof RegExp ||
                typeof fmtDef.compare != "function") {
                throw new Error(`"${keyword***REMOVED***": format "${format***REMOVED***" does not define "compare" function`);
            ***REMOVED***
            const fmt = gen.scopeValue("formats", {
                key: format,
                ref: fmtDef,
                code: opts.code.formats ? codegen_1._ `${opts.code.formats***REMOVED***${codegen_1.getProperty(format)***REMOVED***` : undefined,
            ***REMOVED***);
            cxt.fail$data(compareCode(fmt));
        ***REMOVED***
        function compareCode(fmt) {
            return codegen_1._ `${fmt***REMOVED***.compare(${data***REMOVED***, ${schemaCode***REMOVED***) ${KWDs[keyword].fail***REMOVED*** 0`;
        ***REMOVED***
    ***REMOVED***,
    dependencies: ["format"],
***REMOVED***;
const formatLimitPlugin = (ajv) => {
    ajv.addKeyword(exports.formatLimitDefinition);
    return ajv;
***REMOVED***;
exports.default = formatLimitPlugin;
