// path-type/index.js
'use strict';
const {promisify***REMOVED*** = require('util');
const fs = require('fs');

async function isType(fsStatType, statsMethodName, filePath) {
	if (typeof filePath !== 'string') {
		throw new TypeError(`Expected a string, got ${typeof filePath***REMOVED***`);
	***REMOVED***

	try {
		const stats = await promisify(fs[fsStatType])(filePath);
		return stats[statsMethodName]();
	***REMOVED*** catch (error) {
		if (error.code === 'ENOENT') {
			return false;
		***REMOVED***

		throw error;
	***REMOVED***
***REMOVED***

function isTypeSync(fsStatType, statsMethodName, filePath) {
	if (typeof filePath !== 'string') {
		throw new TypeError(`Expected a string, got ${typeof filePath***REMOVED***`);
	***REMOVED***

	try {
		return fs[fsStatType](filePath)[statsMethodName]();
	***REMOVED*** catch (error) {
		if (error.code === 'ENOENT') {
			return false;
		***REMOVED***

		throw error;
	***REMOVED***
***REMOVED***

exports.isFile = isType.bind(null, 'stat', 'isFile');
exports.isDirectory = isType.bind(null, 'stat', 'isDirectory');
exports.isSymlink = isType.bind(null, 'lstat', 'isSymbolicLink');
exports.isFileSync = isTypeSync.bind(null, 'statSync', 'isFile');
exports.isDirectorySync = isTypeSync.bind(null, 'statSync', 'isDirectory');
exports.isSymlinkSync = isTypeSync.bind(null, 'lstatSync', 'isSymbolicLink');
