// sucrase/dist/HelperManager.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true***REMOVED***);

const HELPERS = {
  require: `
    import {createRequire as CREATE_REQUIRE_NAME***REMOVED*** from "module";
    const require = CREATE_REQUIRE_NAME(import.meta.url);
  `,
  interopRequireWildcard: `
    function interopRequireWildcard(obj) {
      if (obj && obj.__esModule) {
        return obj;
      ***REMOVED*** else {
        var newObj = {***REMOVED***;
        if (obj != null) {
          for (var key in obj) {
            if (Object.prototype.hasOwnProperty.call(obj, key)) {
              newObj[key] = obj[key];
            ***REMOVED***
          ***REMOVED***
        ***REMOVED***
        newObj.default = obj;
        return newObj;
      ***REMOVED***
    ***REMOVED***
  `,
  interopRequireDefault: `
    function interopRequireDefault(obj) {
      return obj && obj.__esModule ? obj : { default: obj ***REMOVED***;
    ***REMOVED***
  `,
  createNamedExportFrom: `
    function createNamedExportFrom(obj, localName, importedName) {
      Object.defineProperty(exports, localName, {enumerable: true, configurable: true, get: () => obj[importedName]***REMOVED***);
    ***REMOVED***
  `,
  // Note that TypeScript and Babel do this differently; TypeScript does a simple existence
  // check in the exports object and does a plain assignment, whereas Babel uses
  // defineProperty and builds an object of explicitly-exported names so that star exports can
  // always take lower precedence. For now, we do the easier TypeScript thing.
  createStarExport: `
    function createStarExport(obj) {
      Object.keys(obj)
        .filter((key) => key !== "default" && key !== "__esModule")
        .forEach((key) => {
          if (exports.hasOwnProperty(key)) {
            return;
          ***REMOVED***
          Object.defineProperty(exports, key, {enumerable: true, configurable: true, get: () => obj[key]***REMOVED***);
        ***REMOVED***);
    ***REMOVED***
  `,
  nullishCoalesce: `
    function nullishCoalesce(lhs, rhsFn) {
      if (lhs != null) {
        return lhs;
      ***REMOVED*** else {
        return rhsFn();
      ***REMOVED***
    ***REMOVED***
  `,
  asyncNullishCoalesce: `
    async function asyncNullishCoalesce(lhs, rhsFn) {
      if (lhs != null) {
        return lhs;
      ***REMOVED*** else {
        return await rhsFn();
      ***REMOVED***
    ***REMOVED***
  `,
  optionalChain: `
    function optionalChain(ops) {
      let lastAccessLHS = undefined;
      let value = ops[0];
      let i = 1;
      while (i < ops.length) {
        const op = ops[i];
        const fn = ops[i + 1];
        i += 2;
        if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) {
          return undefined;
        ***REMOVED***
        if (op === 'access' || op === 'optionalAccess') {
          lastAccessLHS = value;
          value = fn(value);
        ***REMOVED*** else if (op === 'call' || op === 'optionalCall') {
          value = fn((...args) => value.call(lastAccessLHS, ...args));
          lastAccessLHS = undefined;
        ***REMOVED***
      ***REMOVED***
      return value;
    ***REMOVED***
  `,
  asyncOptionalChain: `
    async function asyncOptionalChain(ops) {
      let lastAccessLHS = undefined;
      let value = ops[0];
      let i = 1;
      while (i < ops.length) {
        const op = ops[i];
        const fn = ops[i + 1];
        i += 2;
        if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) {
          return undefined;
        ***REMOVED***
        if (op === 'access' || op === 'optionalAccess') {
          lastAccessLHS = value;
          value = await fn(value);
        ***REMOVED*** else if (op === 'call' || op === 'optionalCall') {
          value = await fn((...args) => value.call(lastAccessLHS, ...args));
          lastAccessLHS = undefined;
        ***REMOVED***
      ***REMOVED***
      return value;
    ***REMOVED***
  `,
  optionalChainDelete: `
    function optionalChainDelete(ops) {
      const result = OPTIONAL_CHAIN_NAME(ops);
      return result == null ? true : result;
    ***REMOVED***
  `,
  asyncOptionalChainDelete: `
    async function asyncOptionalChainDelete(ops) {
      const result = await ASYNC_OPTIONAL_CHAIN_NAME(ops);
      return result == null ? true : result;
    ***REMOVED***
  `,
***REMOVED***;

 class HelperManager {
  __init() {this.helperNames = {***REMOVED******REMOVED***
  __init2() {this.createRequireName = null***REMOVED***
  constructor( nameManager) {;this.nameManager = nameManager;HelperManager.prototype.__init.call(this);HelperManager.prototype.__init2.call(this);***REMOVED***

  getHelperName(baseName) {
    let helperName = this.helperNames[baseName];
    if (helperName) {
      return helperName;
    ***REMOVED***
    helperName = this.nameManager.claimFreeName(`_${baseName***REMOVED***`);
    this.helperNames[baseName] = helperName;
    return helperName;
  ***REMOVED***

  emitHelpers() {
    let resultCode = "";
    if (this.helperNames.optionalChainDelete) {
      this.getHelperName("optionalChain");
    ***REMOVED***
    if (this.helperNames.asyncOptionalChainDelete) {
      this.getHelperName("asyncOptionalChain");
    ***REMOVED***
    for (const [baseName, helperCodeTemplate] of Object.entries(HELPERS)) {
      const helperName = this.helperNames[baseName];
      let helperCode = helperCodeTemplate;
      if (baseName === "optionalChainDelete") {
        helperCode = helperCode.replace("OPTIONAL_CHAIN_NAME", this.helperNames.optionalChain);
      ***REMOVED*** else if (baseName === "asyncOptionalChainDelete") {
        helperCode = helperCode.replace(
          "ASYNC_OPTIONAL_CHAIN_NAME",
          this.helperNames.asyncOptionalChain,
        );
      ***REMOVED*** else if (baseName === "require") {
        if (this.createRequireName === null) {
          this.createRequireName = this.nameManager.claimFreeName("_createRequire");
        ***REMOVED***
        helperCode = helperCode.replace(/CREATE_REQUIRE_NAME/g, this.createRequireName);
      ***REMOVED***
      if (helperName) {
        resultCode += " ";
        resultCode += helperCode.replace(baseName, helperName).replace(/\s+/g, " ").trim();
      ***REMOVED***
    ***REMOVED***
    return resultCode;
  ***REMOVED***
***REMOVED*** exports.HelperManager = HelperManager;
