// sucrase/dist/register.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true***REMOVED***); function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; ***REMOVED*** else { var newObj = {***REMOVED***; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { newObj[key] = obj[key]; ***REMOVED*** ***REMOVED*** ***REMOVED*** newObj.default = obj; return newObj; ***REMOVED*** ***REMOVED***var _pirates = require('pirates'); var pirates = _interopRequireWildcard(_pirates);

var _index = require('./index');








 function addHook(
  extension,
  sucraseOptions,
  hookOptions,
) {
  let mergedSucraseOptions = sucraseOptions;
  const sucraseOptionsEnvJSON = process.env.SUCRASE_OPTIONS;
  if (sucraseOptionsEnvJSON) {
    mergedSucraseOptions = {...mergedSucraseOptions, ...JSON.parse(sucraseOptionsEnvJSON)***REMOVED***;
  ***REMOVED***
  return pirates.addHook(
    (code, filePath) => {
      const {code: transformedCode, sourceMap***REMOVED*** = _index.transform.call(void 0, code, {
        ...mergedSucraseOptions,
        sourceMapOptions: {compiledFilename: filePath***REMOVED***,
        filePath,
      ***REMOVED***);
      const mapBase64 = Buffer.from(JSON.stringify(sourceMap)).toString("base64");
      const suffix = `//# sourceMappingURL=data:application/json;charset=utf-8;base64,${mapBase64***REMOVED***`;
      return `${transformedCode***REMOVED***\n${suffix***REMOVED***`;
    ***REMOVED***,
    {...hookOptions, exts: [extension]***REMOVED***,
  );
***REMOVED*** exports.addHook = addHook;

 function registerJS(hookOptions) {
  return addHook(".js", {transforms: ["imports", "flow", "jsx"]***REMOVED***, hookOptions);
***REMOVED*** exports.registerJS = registerJS;

 function registerJSX(hookOptions) {
  return addHook(".jsx", {transforms: ["imports", "flow", "jsx"]***REMOVED***, hookOptions);
***REMOVED*** exports.registerJSX = registerJSX;

 function registerTS(hookOptions) {
  return addHook(".ts", {transforms: ["imports", "typescript"]***REMOVED***, hookOptions);
***REMOVED*** exports.registerTS = registerTS;

 function registerTSX(hookOptions) {
  return addHook(".tsx", {transforms: ["imports", "typescript", "jsx"]***REMOVED***, hookOptions);
***REMOVED*** exports.registerTSX = registerTSX;

 function registerTSLegacyModuleInterop(hookOptions) {
  return addHook(
    ".ts",
    {
      transforms: ["imports", "typescript"],
      enableLegacyTypeScriptModuleInterop: true,
    ***REMOVED***,
    hookOptions,
  );
***REMOVED*** exports.registerTSLegacyModuleInterop = registerTSLegacyModuleInterop;

 function registerTSXLegacyModuleInterop(hookOptions) {
  return addHook(
    ".tsx",
    {
      transforms: ["imports", "typescript", "jsx"],
      enableLegacyTypeScriptModuleInterop: true,
    ***REMOVED***,
    hookOptions,
  );
***REMOVED*** exports.registerTSXLegacyModuleInterop = registerTSXLegacyModuleInterop;

 function registerAll(hookOptions) {
  const reverts = [
    registerJS(hookOptions),
    registerJSX(hookOptions),
    registerTS(hookOptions),
    registerTSX(hookOptions),
  ];

  return () => {
    for (const fn of reverts) {
      fn();
    ***REMOVED***
  ***REMOVED***;
***REMOVED*** exports.registerAll = registerAll;
