// eslint-plugin-jsx-a11y/__tests__/__util__/parserOptionsMapper.js
import { version as eslintVersion ***REMOVED*** from 'eslint/package.json';
import semver from 'semver';

const usingLegacy = semver.major(eslintVersion) < 9;

const defaultParserOptions = {
  ecmaFeatures: {
    experimentalObjectRestSpread: true,
    jsx: true,
  ***REMOVED***,
***REMOVED***;

const defaultLegacyParserOptions = {
  ...defaultParserOptions,
  ecmaVersion: 2018,
***REMOVED***;

const defaultLanguageOptions = {
  ecmaVersion: 'latest',
  parserOptions: {
    ...defaultParserOptions,
  ***REMOVED***,
***REMOVED***;

export default function parserOptionsMapper({
  code,
  errors,
  options = [],
  languageOptions = {***REMOVED***,
  settings = {***REMOVED***,
***REMOVED***) {
  return usingLegacy
    ? {
      code,
      errors,
      options,
      parserOptions: {
        ...defaultLegacyParserOptions,
        ...languageOptions,
      ***REMOVED***,
      settings,
    ***REMOVED***
    : {
      code,
      errors,
      options,
      languageOptions: {
        ...defaultLanguageOptions,
        ...languageOptions,
      ***REMOVED***,
      settings,
    ***REMOVED***;
***REMOVED***
