// html-webpack-plugin/index.js
// @ts-check
"use strict";

const promisify = require("util").promisify;

const vm = require("vm");
const fs = require("fs");
const _uniq = require("lodash/uniq");
const path = require("path");
const { CachedChildCompilation ***REMOVED*** = require("./lib/cached-child-compiler");

const {
  createHtmlTagObject,
  htmlTagObjectToString,
  HtmlTagArray,
***REMOVED*** = require("./lib/html-tags");
const prettyError = require("./lib/errors.js");
const chunkSorter = require("./lib/chunksorter.js");
const { AsyncSeriesWaterfallHook ***REMOVED*** = require("tapable");

/** @typedef {import("./typings").HtmlTagObject***REMOVED*** HtmlTagObject */
/** @typedef {import("./typings").Options***REMOVED*** HtmlWebpackOptions */
/** @typedef {import("./typings").ProcessedOptions***REMOVED*** ProcessedHtmlWebpackOptions */
/** @typedef {import("./typings").TemplateParameter***REMOVED*** TemplateParameter */
/** @typedef {import("webpack").Compiler***REMOVED*** Compiler */
/** @typedef {import("webpack").Compilation***REMOVED*** Compilation */
/** @typedef {Required<Compilation["outputOptions"]["publicPath"]>***REMOVED*** PublicPath */
/** @typedef {ReturnType<Compiler["getInfrastructureLogger"]>***REMOVED*** Logger */
/** @typedef {Compilation["entrypoints"] extends Map<string, infer I> ? I : never***REMOVED*** Entrypoint */
/** @typedef {Array<{ name: string, source: import('webpack').sources.Source, info?: import('webpack').AssetInfo ***REMOVED***>***REMOVED*** PreviousEmittedAssets */
/** @typedef {{ publicPath: string, js: Array<string>, css: Array<string>, manifest?: string, favicon?: string ***REMOVED******REMOVED*** AssetsInformationByGroups */
/** @typedef {import("./typings").Hooks***REMOVED*** HtmlWebpackPluginHooks */
/**
 * @type {WeakMap<Compilation, HtmlWebpackPluginHooks>***REMOVED******REMOVED***
 */
const compilationHooksMap = new WeakMap();

class HtmlWebpackPlugin {
  // The following is the API definition for all available hooks
  // For the TypeScript definition, see the Hooks type in typings.d.ts
  /**
   beforeAssetTagGeneration:
   AsyncSeriesWaterfallHook<{
      assets: {
        publicPath: string,
        js: Array<string>,
        css: Array<string>,
        favicon?: string | undefined,
        manifest?: string | undefined
      ***REMOVED***,
      outputName: string,
      plugin: HtmlWebpackPlugin
    ***REMOVED***>,
   alterAssetTags:
   AsyncSeriesWaterfallHook<{
      assetTags: {
        scripts: Array<HtmlTagObject>,
        styles: Array<HtmlTagObject>,
        meta: Array<HtmlTagObject>,
      ***REMOVED***,
      publicPath: string,
      outputName: string,
      plugin: HtmlWebpackPlugin
    ***REMOVED***>,
   alterAssetTagGroups:
   AsyncSeriesWaterfallHook<{
      headTags: Array<HtmlTagObject | HtmlTagObject>,
      bodyTags: Array<HtmlTagObject | HtmlTagObject>,
      publicPath: string,
      outputName: string,
      plugin: HtmlWebpackPlugin
    ***REMOVED***>,
   afterTemplateExecution:
   AsyncSeriesWaterfallHook<{
      html: string,
      headTags: Array<HtmlTagObject | HtmlTagObject>,
      bodyTags: Array<HtmlTagObject | HtmlTagObject>,
      outputName: string,
      plugin: HtmlWebpackPlugin,
    ***REMOVED***>,
   beforeEmit:
   AsyncSeriesWaterfallHook<{
      html: string,
      outputName: string,
      plugin: HtmlWebpackPlugin,
    ***REMOVED***>,
   afterEmit:
   AsyncSeriesWaterfallHook<{
      outputName: string,
      plugin: HtmlWebpackPlugin
    ***REMOVED***>
   */

  /**
   * Returns all public hooks of the html webpack plugin for the given compilation
   *
   * @param {Compilation***REMOVED*** compilation
   * @returns {HtmlWebpackPluginHooks***REMOVED***
   */
  static getCompilationHooks(compilation) {
    let hooks = compilationHooksMap.get(compilation);

    if (!hooks) {
      hooks = {
        beforeAssetTagGeneration: new AsyncSeriesWaterfallHook(["pluginArgs"]),
        alterAssetTags: new AsyncSeriesWaterfallHook(["pluginArgs"]),
        alterAssetTagGroups: new AsyncSeriesWaterfallHook(["pluginArgs"]),
        afterTemplateExecution: new AsyncSeriesWaterfallHook(["pluginArgs"]),
        beforeEmit: new AsyncSeriesWaterfallHook(["pluginArgs"]),
        afterEmit: new AsyncSeriesWaterfallHook(["pluginArgs"]),
      ***REMOVED***;
      compilationHooksMap.set(compilation, hooks);
    ***REMOVED***

    return hooks;
  ***REMOVED***

  /**
   * @param {HtmlWebpackOptions***REMOVED*** [options]
   */
  constructor(options) {
    /** @type {HtmlWebpackOptions***REMOVED*** */
    // TODO remove me in the next major release
    this.userOptions = options || {***REMOVED***;
    this.version = HtmlWebpackPlugin.version;

    // Default options
    /** @type {ProcessedHtmlWebpackOptions***REMOVED*** */
    const defaultOptions = {
      template: "auto",
      templateContent: false,
      templateParameters: templateParametersGenerator,
      filename: "index.html",
      publicPath:
        this.userOptions.publicPath === undefined
          ? "auto"
          : this.userOptions.publicPath,
      hash: false,
      inject: this.userOptions.scriptLoading === "blocking" ? "body" : "head",
      scriptLoading: "defer",
      compile: true,
      favicon: false,
      minify: "auto",
      cache: true,
      showErrors: true,
      chunks: "all",
      excludeChunks: [],
      chunksSortMode: "auto",
      meta: {***REMOVED***,
      base: false,
      title: "Webpack App",
      xhtml: false,
    ***REMOVED***;

    /** @type {ProcessedHtmlWebpackOptions***REMOVED*** */
    this.options = Object.assign(defaultOptions, this.userOptions);
  ***REMOVED***

  /**
   *
   * @param {Compiler***REMOVED*** compiler
   * @returns {void***REMOVED***
   */
  apply(compiler) {
    this.logger = compiler.getInfrastructureLogger("HtmlWebpackPlugin");

    const options = this.options;

    options.template = this.getTemplatePath(
      this.options.template,
      compiler.context,
    );

    // Assert correct option spelling
    if (
      options.scriptLoading !== "defer" &&
      options.scriptLoading !== "blocking" &&
      options.scriptLoading !== "module" &&
      options.scriptLoading !== "systemjs-module"
    ) {
      /** @type {Logger***REMOVED*** */
      (this.logger).error(
        'The "scriptLoading" option need to be set to "defer", "blocking" or "module" or "systemjs-module"',
      );
    ***REMOVED***

    if (
      options.inject !== true &&
      options.inject !== false &&
      options.inject !== "head" &&
      options.inject !== "body"
    ) {
      /** @type {Logger***REMOVED*** */
      (this.logger).error(
        'The `inject` option needs to be set to true, false, "head" or "body',
      );
    ***REMOVED***

    if (
      this.options.templateParameters !== false &&
      typeof this.options.templateParameters !== "function" &&
      typeof this.options.templateParameters !== "object"
    ) {
      /** @type {Logger***REMOVED*** */
      (this.logger).error(
        "The `templateParameters` has to be either a function or an object or false",
      );
    ***REMOVED***

    // Default metaOptions if no template is provided
    if (
      !this.userOptions.template &&
      options.templateContent === false &&
      options.meta
    ) {
      options.meta = Object.assign(
        {***REMOVED***,
        options.meta,
        {
          // TODO remove in the next major release
          // From https://developer.mozilla.org/en-US/docs/Mozilla/Mobile/Viewport_meta_tag
          viewport: "width=device-width, initial-scale=1",
        ***REMOVED***,
        this.userOptions.meta,
      );
    ***REMOVED***

    // entryName to fileName conversion function
    const userOptionFilename =
      this.userOptions.filename || this.options.filename;
    const filenameFunction =
      typeof userOptionFilename === "function"
        ? userOptionFilename
        : // Replace '[name]' with entry name
          (entryName) => userOptionFilename.replace(/\[name\]/g, entryName);

    /** output filenames for the given entry names */
    const entryNames = Object.keys(compiler.options.entry);
    const outputFileNames = new Set(
      (entryNames.length ? entryNames : ["main"]).map(filenameFunction),
    );

    // Hook all options into the webpack compiler
    outputFileNames.forEach((outputFileName) => {
      // Instance variables to keep caching information for multiple builds
      const assetJson = { value: undefined ***REMOVED***;
      /**
       * store the previous generated asset to emit them even if the content did not change
       * to support watch mode for third party plugins like the clean-webpack-plugin or the compression plugin
       * @type {PreviousEmittedAssets***REMOVED***
       */
      const previousEmittedAssets = [];

      // Inject child compiler plugin
      const childCompilerPlugin = new CachedChildCompilation(compiler);

      if (!this.options.templateContent) {
        childCompilerPlugin.addEntry(this.options.template);
      ***REMOVED***

      // convert absolute filename into relative so that webpack can
      // generate it at correct location
      let filename = outputFileName;

      if (path.resolve(filename) === path.normalize(filename)) {
        const outputPath =
          /** @type {string***REMOVED*** - Once initialized the path is always a string */ (
            compiler.options.output.path
          );

        filename = path.relative(outputPath, filename);
      ***REMOVED***

      compiler.hooks.thisCompilation.tap(
        "HtmlWebpackPlugin",
        /**
         * Hook into the webpack compilation
         * @param {Compilation***REMOVED*** compilation
         */
        (compilation) => {
          compilation.hooks.processAssets.tapAsync(
            {
              name: "HtmlWebpackPlugin",
              stage:
                /**
                 * Generate the html after minification and dev tooling is done
                 */
                compiler.webpack.Compilation
                  .PROCESS_ASSETS_STAGE_OPTIMIZE_INLINE,
            ***REMOVED***,
            /**
             * Hook into the process assets hook
             * @param {any***REMOVED*** _
             * @param {(err?: Error) => void***REMOVED*** callback
             */
            (_, callback) => {
              this.generateHTML(
                compiler,
                compilation,
                filename,
                childCompilerPlugin,
                previousEmittedAssets,
                assetJson,
                callback,
              );
            ***REMOVED***,
          );
        ***REMOVED***,
      );
    ***REMOVED***);
  ***REMOVED***

  /**
   * Helper to return the absolute template path with a fallback loader
   *
   * @private
   * @param {string***REMOVED*** template The path to the template e.g. './index.html'
   * @param {string***REMOVED*** context The webpack base resolution path for relative paths e.g. process.cwd()
   */
  getTemplatePath(template, context) {
    if (template === "auto") {
      template = path.resolve(context, "src/index.ejs");
      if (!fs.existsSync(template)) {
        template = path.join(__dirname, "default_index.ejs");
      ***REMOVED***
    ***REMOVED***

    // If the template doesn't use a loader use the lodash template loader
    if (template.indexOf("!") === -1) {
      template =
        require.resolve("./lib/loader.js") +
        "!" +
        path.resolve(context, template);
    ***REMOVED***

    // Resolve template path
    return template.replace(
      /([!])([^/\\][^!?]+|[^/\\!?])($|\?[^!?\n]+$)/,
      (match, prefix, filepath, postfix) =>
        prefix + path.resolve(filepath) + postfix,
    );
  ***REMOVED***

  /**
   * Return all chunks from the compilation result which match the exclude and include filters
   *
   * @private
   * @param {any***REMOVED*** chunks
   * @param {string[]|'all'***REMOVED*** includedChunks
   * @param {string[]***REMOVED*** excludedChunks
   */
  filterEntryChunks(chunks, includedChunks, excludedChunks) {
    return chunks.filter((chunkName) => {
      // Skip if the chunks should be filtered and the given chunk was not added explicity
      if (
        Array.isArray(includedChunks) &&
        includedChunks.indexOf(chunkName) === -1
      ) {
        return false;
      ***REMOVED***

      // Skip if the chunks should be filtered and the given chunk was excluded explicity
      if (
        Array.isArray(excludedChunks) &&
        excludedChunks.indexOf(chunkName) !== -1
      ) {
        return false;
      ***REMOVED***

      // Add otherwise
      return true;
    ***REMOVED***);
  ***REMOVED***

  /**
   * Helper to sort chunks
   *
   * @private
   * @param {string[]***REMOVED*** entryNames
   * @param {string|((entryNameA: string, entryNameB: string) => number)***REMOVED*** sortMode
   * @param {Compilation***REMOVED*** compilation
   */
  sortEntryChunks(entryNames, sortMode, compilation) {
    // Custom function
    if (typeof sortMode === "function") {
      return entryNames.sort(sortMode);
    ***REMOVED***
    // Check if the given sort mode is a valid chunkSorter sort mode
    if (typeof chunkSorter[sortMode] !== "undefined") {
      return chunkSorter[sortMode](entryNames, compilation, this.options);
    ***REMOVED***
    throw new Error('"' + sortMode + '" is not a valid chunk sort mode');
  ***REMOVED***

  /**
   * Encode each path component using `encodeURIComponent` as files can contain characters
   * which needs special encoding in URLs like `+ `.
   *
   * Valid filesystem characters which need to be encoded for urls:
   *
   * # pound, % percent, & ampersand, { left curly bracket, ***REMOVED*** right curly bracket,
   * \ back slash, < left angle bracket, > right angle bracket, * asterisk, ? question mark,
   * blank spaces, $ dollar sign, ! exclamation point, ' single quotes, " double quotes,
   * : colon, @ at sign, + plus sign, ` backtick, | pipe, = equal sign
   *
   * However the query string must not be encoded:
   *
   *  fo:demonstration-path/very fancy+name.js?path=/home?value=abc&value=def#zzz
   *    ^             ^    ^    ^     ^    ^  ^    ^^    ^     ^   ^     ^   ^
   *    |             |    |    |     |    |  |    ||    |     |   |     |   |
   *    encoded       |    |    encoded    |  |    ||    |     |   |     |   |
   *                 ignored              ignored  ignored     ignored   ignored
   *
   * @private
   * @param {string***REMOVED*** filePath
   */
  urlencodePath(filePath) {
    // People use the filepath in quite unexpected ways.
    // Try to extract the first querystring of the url:
    //
    // some+path/demo.html?value=abc?def
    //
    const queryStringStart = filePath.indexOf("?");
    const urlPath =
      queryStringStart === -1 ? filePath : filePath.substr(0, queryStringStart);
    const queryString = filePath.substr(urlPath.length);
    // Encode all parts except '/' which are not part of the querystring:
    const encodedUrlPath = urlPath.split("/").map(encodeURIComponent).join("/");
    return encodedUrlPath + queryString;
  ***REMOVED***

  /**
   * Appends a cache busting hash to the query string of the url
   * E.g. http://localhost:8080/ -> http://localhost:8080/?50c9096ba6183fd728eeb065a26ec175
   *
   * @private
   * @param {string | undefined***REMOVED*** url
   * @param {string***REMOVED*** hash
   */
  appendHash(url, hash) {
    if (!url) {
      return url;
    ***REMOVED***

    return url + (url.indexOf("?") === -1 ? "?" : "&") + hash;
  ***REMOVED***

  /**
   * Generate the relative or absolute base url to reference images, css, and javascript files
   * from within the html file - the publicPath
   *
   * @private
   * @param {Compilation***REMOVED*** compilation
   * @param {string***REMOVED*** filename
   * @param {string | 'auto'***REMOVED*** customPublicPath
   * @returns {string***REMOVED***
   */
  getPublicPath(compilation, filename, customPublicPath) {
    /**
     * @type {string***REMOVED*** the configured public path to the asset root
     * if a path publicPath is set in the current webpack config use it otherwise
     * fallback to a relative path
     */
    const webpackPublicPath = compilation.getAssetPath(
      /** @type {NonNullable<Compilation["outputOptions"]["publicPath"]>***REMOVED*** */ (
        compilation.outputOptions.publicPath
      ),
      { hash: compilation.hash ***REMOVED***,
    );
    // Webpack 5 introduced "auto" as default value
    const isPublicPathDefined = webpackPublicPath !== "auto";

    let publicPath =
      // If the html-webpack-plugin options contain a custom public path unset it
      customPublicPath !== "auto"
        ? customPublicPath
        : isPublicPathDefined
          ? // If a hard coded public path exists use it
            webpackPublicPath
          : // If no public path was set get a relative url path
            path
              .relative(
                path.resolve(
                  /** @type {string***REMOVED*** */ (compilation.options.output.path),
                  path.dirname(filename),
                ),
                /** @type {string***REMOVED*** */ (compilation.options.output.path),
              )
              .split(path.sep)
              .join("/");

    if (publicPath.length && publicPath.substr(-1, 1) !== "/") {
      publicPath += "/";
    ***REMOVED***

    return publicPath;
  ***REMOVED***

  /**
   * The getAssetsForHTML extracts the asset information of a webpack compilation for all given entry names.
   *
   * @private
   * @param {Compilation***REMOVED*** compilation
   * @param {string***REMOVED*** outputName
   * @param {string[]***REMOVED*** entryNames
   * @returns {AssetsInformationByGroups***REMOVED***
   */
  getAssetsInformationByGroups(compilation, outputName, entryNames) {
    /** The public path used inside the html file */
    const publicPath = this.getPublicPath(
      compilation,
      outputName,
      this.options.publicPath,
    );
    /**
     * @type {AssetsInformationByGroups***REMOVED***
     */
    const assets = {
      // The public path
      publicPath,
      // Will contain all js and mjs files
      js: [],
      // Will contain all css files
      css: [],
      // Will contain the html5 appcache manifest files if it exists
      manifest: Object.keys(compilation.assets).find(
        (assetFile) => path.extname(assetFile) === ".appcache",
      ),
      // Favicon
      favicon: undefined,
    ***REMOVED***;

    // Append a hash for cache busting
    if (this.options.hash && assets.manifest) {
      assets.manifest = this.appendHash(
        assets.manifest,
        /** @type {string***REMOVED*** */ (compilation.hash),
      );
    ***REMOVED***

    // Extract paths to .js, .mjs and .css files from the current compilation
    const entryPointPublicPathMap = {***REMOVED***;
    const extensionRegexp = /\.(css|js|mjs)(\?|$)/;

    for (let i = 0; i < entryNames.length; i++) {
      const entryName = entryNames[i];
      /** entryPointUnfilteredFiles - also includes hot module update files */
      const entryPointUnfilteredFiles = /** @type {Entrypoint***REMOVED*** */ (
        compilation.entrypoints.get(entryName)
      ).getFiles();
      const entryPointFiles = entryPointUnfilteredFiles.filter((chunkFile) => {
        const asset = compilation.getAsset(chunkFile);

        if (!asset) {
          return true;
        ***REMOVED***

        // Prevent hot-module files from being included:
        const assetMetaInformation = asset.info || {***REMOVED***;

        return !(
          assetMetaInformation.hotModuleReplacement ||
          assetMetaInformation.development
        );
      ***REMOVED***);
      // Prepend the publicPath and append the hash depending on the
      // webpack.output.publicPath and hashOptions
      // E.g. bundle.js -> /bundle.js?hash
      const entryPointPublicPaths = entryPointFiles.map((chunkFile) => {
        const entryPointPublicPath = publicPath + this.urlencodePath(chunkFile);
        return this.options.hash
          ? this.appendHash(
              entryPointPublicPath,
              /** @type {string***REMOVED*** */ (compilation.hash),
            )
          : entryPointPublicPath;
      ***REMOVED***);

      entryPointPublicPaths.forEach((entryPointPublicPath) => {
        const extMatch = extensionRegexp.exec(
          /** @type {string***REMOVED*** */ (entryPointPublicPath),
        );

        // Skip if the public path is not a .css, .mjs or .js file
        if (!extMatch) {
          return;
        ***REMOVED***

        // Skip if this file is already known
        // (e.g. because of common chunk optimizations)
        if (entryPointPublicPathMap[entryPointPublicPath]) {
          return;
        ***REMOVED***

        entryPointPublicPathMap[entryPointPublicPath] = true;

        // ext will contain .js or .css, because .mjs recognizes as .js
        const ext = extMatch[1] === "mjs" ? "js" : extMatch[1];

        assets[ext].push(entryPointPublicPath);
      ***REMOVED***);
    ***REMOVED***

    return assets;
  ***REMOVED***

  /**
   * Once webpack is done with compiling the template into a NodeJS code this function
   * evaluates it to generate the html result
   *
   * The evaluateCompilationResult is only a class function to allow spying during testing.
   * Please change that in a further refactoring
   *
   * @param {string***REMOVED*** source
   * @param {string***REMOVED*** publicPath
   * @param {string***REMOVED*** templateFilename
   * @returns {Promise<string | (() => string | Promise<string>)>***REMOVED***
   */
  evaluateCompilationResult(source, publicPath, templateFilename) {
    if (!source) {
      return Promise.reject(
        new Error("The child compilation didn't provide a result"),
      );
    ***REMOVED***

    // The LibraryTemplatePlugin stores the template result in a local variable.
    // By adding it to the end the value gets extracted during evaluation
    if (source.indexOf("HTML_WEBPACK_PLUGIN_RESULT") >= 0) {
      source += ";\nHTML_WEBPACK_PLUGIN_RESULT";
    ***REMOVED***

    const templateWithoutLoaders = templateFilename
      .replace(/^.+!/, "")
      .replace(/\?.+$/, "");
    const vmContext = vm.createContext({
      ...global,
      HTML_WEBPACK_PLUGIN: true,
      require: require,
      htmlWebpackPluginPublicPath: publicPath,
      __filename: templateWithoutLoaders,
      __dirname: path.dirname(templateWithoutLoaders),
      AbortController: global.AbortController,
      AbortSignal: global.AbortSignal,
      Blob: global.Blob,
      Buffer: global.Buffer,
      ByteLengthQueuingStrategy: global.ByteLengthQueuingStrategy,
      BroadcastChannel: global.BroadcastChannel,
      CompressionStream: global.CompressionStream,
      CountQueuingStrategy: global.CountQueuingStrategy,
      Crypto: global.Crypto,
      CryptoKey: global.CryptoKey,
      CustomEvent: global.CustomEvent,
      DecompressionStream: global.DecompressionStream,
      Event: global.Event,
      EventTarget: global.EventTarget,
      File: global.File,
      FormData: global.FormData,
      Headers: global.Headers,
      MessageChannel: global.MessageChannel,
      MessageEvent: global.MessageEvent,
      MessagePort: global.MessagePort,
      PerformanceEntry: global.PerformanceEntry,
      PerformanceMark: global.PerformanceMark,
      PerformanceMeasure: global.PerformanceMeasure,
      PerformanceObserver: global.PerformanceObserver,
      PerformanceObserverEntryList: global.PerformanceObserverEntryList,
      PerformanceResourceTiming: global.PerformanceResourceTiming,
      ReadableByteStreamController: global.ReadableByteStreamController,
      ReadableStream: global.ReadableStream,
      ReadableStreamBYOBReader: global.ReadableStreamBYOBReader,
      ReadableStreamBYOBRequest: global.ReadableStreamBYOBRequest,
      ReadableStreamDefaultController: global.ReadableStreamDefaultController,
      ReadableStreamDefaultReader: global.ReadableStreamDefaultReader,
      Response: global.Response,
      Request: global.Request,
      SubtleCrypto: global.SubtleCrypto,
      DOMException: global.DOMException,
      TextDecoder: global.TextDecoder,
      TextDecoderStream: global.TextDecoderStream,
      TextEncoder: global.TextEncoder,
      TextEncoderStream: global.TextEncoderStream,
      TransformStream: global.TransformStream,
      TransformStreamDefaultController: global.TransformStreamDefaultController,
      URL: global.URL,
      URLSearchParams: global.URLSearchParams,
      WebAssembly: global.WebAssembly,
      WritableStream: global.WritableStream,
      WritableStreamDefaultController: global.WritableStreamDefaultController,
      WritableStreamDefaultWriter: global.WritableStreamDefaultWriter,
    ***REMOVED***);

    const vmScript = new vm.Script(source, {
      filename: templateWithoutLoaders,
    ***REMOVED***);

    // Evaluate code and cast to string
    let newSource;

    try {
      newSource = vmScript.runInContext(vmContext);
    ***REMOVED*** catch (e) {
      return Promise.reject(e);
    ***REMOVED***

    if (
      typeof newSource === "object" &&
      newSource.__esModule &&
      newSource.default !== undefined
    ) {
      newSource = newSource.default;
    ***REMOVED***

    return typeof newSource === "string" || typeof newSource === "function"
      ? Promise.resolve(newSource)
      : Promise.reject(
          new Error(
            'The loader "' + templateWithoutLoaders + "\" didn't return html.",
          ),
        );
  ***REMOVED***

  /**
   * Add toString methods for easier rendering inside the template
   *
   * @private
   * @param {Array<HtmlTagObject>***REMOVED*** assetTagGroup
   * @returns {Array<HtmlTagObject>***REMOVED***
   */
  prepareAssetTagGroupForRendering(assetTagGroup) {
    const xhtml = this.options.xhtml;
    return HtmlTagArray.from(
      assetTagGroup.map((assetTag) => {
        const copiedAssetTag = Object.assign({***REMOVED***, assetTag);
        copiedAssetTag.toString = function () {
          return htmlTagObjectToString(this, xhtml);
        ***REMOVED***;
        return copiedAssetTag;
      ***REMOVED***),
    );
  ***REMOVED***

  /**
   * Generate the template parameters for the template function
   *
   * @private
   * @param {Compilation***REMOVED*** compilation
   * @param {AssetsInformationByGroups***REMOVED*** assetsInformationByGroups
   * @param {{
       headTags: HtmlTagObject[],
       bodyTags: HtmlTagObject[]
     ***REMOVED******REMOVED*** assetTags
   * @returns {Promise<{[key: any]: any***REMOVED***>***REMOVED***
   */
  getTemplateParameters(compilation, assetsInformationByGroups, assetTags) {
    const templateParameters = this.options.templateParameters;

    if (templateParameters === false) {
      return Promise.resolve({***REMOVED***);
    ***REMOVED***

    if (
      typeof templateParameters !== "function" &&
      typeof templateParameters !== "object"
    ) {
      throw new Error(
        "templateParameters has to be either a function or an object",
      );
    ***REMOVED***

    const templateParameterFunction =
      typeof templateParameters === "function"
        ? // A custom function can overwrite the entire template parameter preparation
          templateParameters
        : // If the template parameters is an object merge it with the default values
          (compilation, assetsInformationByGroups, assetTags, options) =>
            Object.assign(
              {***REMOVED***,
              templateParametersGenerator(
                compilation,
                assetsInformationByGroups,
                assetTags,
                options,
              ),
              templateParameters,
            );
    const preparedAssetTags = {
      headTags: this.prepareAssetTagGroupForRendering(assetTags.headTags),
      bodyTags: this.prepareAssetTagGroupForRendering(assetTags.bodyTags),
    ***REMOVED***;
    return Promise.resolve().then(() =>
      templateParameterFunction(
        compilation,
        assetsInformationByGroups,
        preparedAssetTags,
        this.options,
      ),
    );
  ***REMOVED***

  /**
   * This function renders the actual html by executing the template function
   *
   * @private
   * @param {(templateParameters) => string | Promise<string>***REMOVED*** templateFunction
   * @param {AssetsInformationByGroups***REMOVED*** assetsInformationByGroups
   * @param {{
       headTags: HtmlTagObject[],
       bodyTags: HtmlTagObject[]
     ***REMOVED******REMOVED*** assetTags
   * @param {Compilation***REMOVED*** compilation
   * @returns Promise<string>
   */
  executeTemplate(
    templateFunction,
    assetsInformationByGroups,
    assetTags,
    compilation,
  ) {
    // Template processing
    const templateParamsPromise = this.getTemplateParameters(
      compilation,
      assetsInformationByGroups,
      assetTags,
    );

    return templateParamsPromise.then((templateParams) => {
      try {
        // If html is a promise return the promise
        // If html is a string turn it into a promise
        return templateFunction(templateParams);
      ***REMOVED*** catch (e) {
        // @ts-ignore
        compilation.errors.push(new Error("Template execution failed: " + e));
        return Promise.reject(e);
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***

  /**
   * Html Post processing
   *
   * @private
   * @param {Compiler***REMOVED*** compiler The compiler instance
   * @param {any***REMOVED*** originalHtml The input html
   * @param {AssetsInformationByGroups***REMOVED*** assetsInformationByGroups
   * @param {{headTags: HtmlTagObject[], bodyTags: HtmlTagObject[]***REMOVED******REMOVED*** assetTags The asset tags to inject
   * @returns {Promise<string>***REMOVED***
   */
  postProcessHtml(
    compiler,
    originalHtml,
    assetsInformationByGroups,
    assetTags,
  ) {
    let html = originalHtml;

    if (typeof html !== "string") {
      return Promise.reject(
        new Error(
          "Expected html to be a string but got " + JSON.stringify(html),
        ),
      );
    ***REMOVED***

    if (this.options.inject) {
      const htmlRegExp = /(<html[^>]*>)/i;
      const headRegExp = /(<\/head\s*>)/i;
      const bodyRegExp = /(<\/body\s*>)/i;
      const metaViewportRegExp = /<meta[^>]+name=["']viewport["'][^>]*>/i;
      const body = assetTags.bodyTags.map((assetTagObject) =>
        htmlTagObjectToString(assetTagObject, this.options.xhtml),
      );
      const head = assetTags.headTags
        .filter((item) => {
          if (
            item.tagName === "meta" &&
            item.attributes &&
            item.attributes.name === "viewport" &&
            metaViewportRegExp.test(html)
          ) {
            return false;
          ***REMOVED***

          return true;
        ***REMOVED***)
        .map((assetTagObject) =>
          htmlTagObjectToString(assetTagObject, this.options.xhtml),
        );

      if (body.length) {
        if (bodyRegExp.test(html)) {
          // Append assets to body element
          html = html.replace(bodyRegExp, (match) => body.join("") + match);
        ***REMOVED*** else {
          // Append scripts to the end of the file if no <body> element exists:
          html += body.join("");
        ***REMOVED***
      ***REMOVED***

      if (head.length) {
        // Create a head tag if none exists
        if (!headRegExp.test(html)) {
          if (!htmlRegExp.test(html)) {
            html = "<head></head>" + html;
          ***REMOVED*** else {
            html = html.replace(htmlRegExp, (match) => match + "<head></head>");
          ***REMOVED***
        ***REMOVED***

        // Append assets to head element
        html = html.replace(headRegExp, (match) => head.join("") + match);
      ***REMOVED***

      // Inject manifest into the opening html tag
      if (assetsInformationByGroups.manifest) {
        html = html.replace(/(<html[^>]*)(>)/i, (match, start, end) => {
          // Append the manifest only if no manifest was specified
          if (/\smanifest\s*=/.test(match)) {
            return match;
          ***REMOVED***
          return (
            start +
            ' manifest="' +
            assetsInformationByGroups.manifest +
            '"' +
            end
          );
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***

    // TODO avoid this logic and use https://github.com/webpack-contrib/html-minimizer-webpack-plugin under the hood in the next major version
    // Check if webpack is running in production mode
    // @see https://github.com/webpack/webpack/blob/3366421f1784c449f415cda5930a8e445086f688/lib/WebpackOptionsDefaulter.js#L12-L14
    const isProductionLikeMode =
      compiler.options.mode === "production" || !compiler.options.mode;
    const needMinify =
      this.options.minify === true ||
      typeof this.options.minify === "object" ||
      (this.options.minify === "auto" && isProductionLikeMode);

    if (!needMinify) {
      return Promise.resolve(html);
    ***REMOVED***

    const minifyOptions =
      typeof this.options.minify === "object"
        ? this.options.minify
        : {
            // https://www.npmjs.com/package/html-minifier-terser#options-quick-reference
            collapseWhitespace: true,
            keepClosingSlash: true,
            removeComments: true,
            removeRedundantAttributes: true,
            removeScriptTypeAttributes: true,
            removeStyleLinkTypeAttributes: true,
            useShortDoctype: true,
          ***REMOVED***;

    try {
      html = require("html-minifier-terser").minify(html, minifyOptions);
    ***REMOVED*** catch (e) {
      const isParseError = String(e.message).indexOf("Parse Error") === 0;

      if (isParseError) {
        e.message =
          "html-webpack-plugin could not minify the generated output.\n" +
          "In production mode the html minification is enabled by default.\n" +
          "If you are not generating a valid html output please disable it manually.\n" +
          "You can do so by adding the following setting to your HtmlWebpackPlugin config:\n|\n|" +
          "    minify: false\n|\n" +
          "See https://github.com/jantimon/html-webpack-plugin#options for details.\n\n" +
          "For parser dedicated bugs please create an issue here:\n" +
          "https://danielruf.github.io/html-minifier-terser/" +
          "\n" +
          e.message;
      ***REMOVED***

      return Promise.reject(e);
    ***REMOVED***

    return Promise.resolve(html);
  ***REMOVED***

  /**
   * Helper to return a sorted unique array of all asset files out of the asset object
   * @private
   */
  getAssetFiles(assets) {
    const files = _uniq(
      Object.keys(assets)
        .filter((assetType) => assetType !== "chunks" && assets[assetType])
        .reduce((files, assetType) => files.concat(assets[assetType]), []),
    );
    files.sort();
    return files;
  ***REMOVED***

  /**
   * Converts a favicon file from disk to a webpack resource and returns the url to the resource
   *
   * @private
   * @param {Compiler***REMOVED*** compiler
   * @param {string|false***REMOVED*** favicon
   * @param {Compilation***REMOVED*** compilation
   * @param {string***REMOVED*** publicPath
   * @param {PreviousEmittedAssets***REMOVED*** previousEmittedAssets
   * @returns {Promise<string|undefined>***REMOVED***
   */
  generateFavicon(
    compiler,
    favicon,
    compilation,
    publicPath,
    previousEmittedAssets,
  ) {
    if (!favicon) {
      return Promise.resolve(undefined);
    ***REMOVED***

    const filename = path.resolve(compilation.compiler.context, favicon);

    return promisify(compilation.inputFileSystem.readFile)(filename)
      .then((buf) => {
        const source = new compiler.webpack.sources.RawSource(
          /** @type {string | Buffer***REMOVED*** */ (buf),
          false,
        );
        const name = path.basename(filename);

        compilation.fileDependencies.add(filename);
        compilation.emitAsset(name, source);
        previousEmittedAssets.push({ name, source ***REMOVED***);

        const faviconPath = publicPath + name;

        if (this.options.hash) {
          return this.appendHash(
            faviconPath,
            /** @type {string***REMOVED*** */ (compilation.hash),
          );
        ***REMOVED***

        return faviconPath;
      ***REMOVED***)
      .catch(() =>
        Promise.reject(
          new Error("HtmlWebpackPlugin: could not load file " + filename),
        ),
      );
  ***REMOVED***

  /**
   * Generate all tags script for the given file paths
   *
   * @private
   * @param {Array<string>***REMOVED*** jsAssets
   * @returns {Array<HtmlTagObject>***REMOVED***
   */
  generatedScriptTags(jsAssets) {
    // @ts-ignore
    return jsAssets.map((src) => {
      const attributes = {***REMOVED***;

      if (this.options.scriptLoading === "defer") {
        attributes.defer = true;
      ***REMOVED*** else if (this.options.scriptLoading === "module") {
        attributes.type = "module";
      ***REMOVED*** else if (this.options.scriptLoading === "systemjs-module") {
        attributes.type = "systemjs-module";
      ***REMOVED***

      attributes.src = src;

      return {
        tagName: "script",
        voidTag: false,
        meta: { plugin: "html-webpack-plugin" ***REMOVED***,
        attributes,
      ***REMOVED***;
    ***REMOVED***);
  ***REMOVED***

  /**
   * Generate all style tags for the given file paths
   *
   * @private
   * @param {Array<string>***REMOVED*** cssAssets
   * @returns {Array<HtmlTagObject>***REMOVED***
   */
  generateStyleTags(cssAssets) {
    return cssAssets.map((styleAsset) => ({
      tagName: "link",
      voidTag: true,
      meta: { plugin: "html-webpack-plugin" ***REMOVED***,
      attributes: {
        href: styleAsset,
        rel: "stylesheet",
      ***REMOVED***,
    ***REMOVED***));
  ***REMOVED***

  /**
   * Generate an optional base tag
   *
   * @param {string | {[attributeName: string]: string***REMOVED******REMOVED*** base
   * @returns {Array<HtmlTagObject>***REMOVED***
   */
  generateBaseTag(base) {
    return [
      {
        tagName: "base",
        voidTag: true,
        meta: { plugin: "html-webpack-plugin" ***REMOVED***,
        // attributes e.g. { href:"http://example.com/page.html" target:"_blank" ***REMOVED***
        attributes:
          typeof base === "string"
            ? {
                href: base,
              ***REMOVED***
            : base,
      ***REMOVED***,
    ];
  ***REMOVED***

  /**
   * Generate all meta tags for the given meta configuration
   *
   * @private
   * @param {false | {[name: string]:  false | string | {[attributeName: string]: string|boolean***REMOVED******REMOVED******REMOVED*** metaOptions
   * @returns {Array<HtmlTagObject>***REMOVED***
   */
  generatedMetaTags(metaOptions) {
    if (metaOptions === false) {
      return [];
    ***REMOVED***

    // Make tags self-closing in case of xhtml
    // Turn { "viewport" : "width=500, initial-scale=1" ***REMOVED*** into
    // [{ name:"viewport" content:"width=500, initial-scale=1" ***REMOVED***]
    const metaTagAttributeObjects = Object.keys(metaOptions)
      .map((metaName) => {
        const metaTagContent = metaOptions[metaName];
        return typeof metaTagContent === "string"
          ? {
              name: metaName,
              content: metaTagContent,
            ***REMOVED***
          : metaTagContent;
      ***REMOVED***)
      .filter((attribute) => attribute !== false);

    // Turn [{ name:"viewport" content:"width=500, initial-scale=1" ***REMOVED***] into
    // the html-webpack-plugin tag structure
    return metaTagAttributeObjects.map((metaTagAttributes) => {
      if (metaTagAttributes === false) {
        throw new Error("Invalid meta tag");
      ***REMOVED***
      return {
        tagName: "meta",
        voidTag: true,
        meta: { plugin: "html-webpack-plugin" ***REMOVED***,
        attributes: metaTagAttributes,
      ***REMOVED***;
    ***REMOVED***);
  ***REMOVED***

  /**
   * Generate a favicon tag for the given file path
   *
   * @private
   * @param {string***REMOVED*** favicon
   * @returns {Array<HtmlTagObject>***REMOVED***
   */
  generateFaviconTag(favicon) {
    return [
      {
        tagName: "link",
        voidTag: true,
        meta: { plugin: "html-webpack-plugin" ***REMOVED***,
        attributes: {
          rel: "icon",
          href: favicon,
        ***REMOVED***,
      ***REMOVED***,
    ];
  ***REMOVED***

  /**
   * Group assets to head and body tags
   *
   * @param {{
      scripts: Array<HtmlTagObject>;
      styles: Array<HtmlTagObject>;
      meta: Array<HtmlTagObject>;
    ***REMOVED******REMOVED*** assetTags
   * @param {"body" | "head"***REMOVED*** scriptTarget
   * @returns {{
      headTags: Array<HtmlTagObject>;
      bodyTags: Array<HtmlTagObject>;
    ***REMOVED******REMOVED***
   */
  groupAssetsByElements(assetTags, scriptTarget) {
    /** @type {{ headTags: Array<HtmlTagObject>; bodyTags: Array<HtmlTagObject>; ***REMOVED******REMOVED*** */
    const result = {
      headTags: [...assetTags.meta, ...assetTags.styles],
      bodyTags: [],
    ***REMOVED***;

    // Add script tags to head or body depending on
    // the htmlPluginOptions
    if (scriptTarget === "body") {
      result.bodyTags.push(...assetTags.scripts);
    ***REMOVED*** else {
      // If script loading is blocking add the scripts to the end of the head
      // If script loading is non-blocking add the scripts in front of the css files
      const insertPosition =
        this.options.scriptLoading === "blocking"
          ? result.headTags.length
          : assetTags.meta.length;

      result.headTags.splice(insertPosition, 0, ...assetTags.scripts);
    ***REMOVED***

    return result;
  ***REMOVED***

  /**
   * Replace [contenthash] in filename
   *
   * @see https://survivejs.com/webpack/optimizing/adding-hashes-to-filenames/
   *
   * @private
   * @param {Compiler***REMOVED*** compiler
   * @param {string***REMOVED*** filename
   * @param {string|Buffer***REMOVED*** fileContent
   * @param {Compilation***REMOVED*** compilation
   * @returns {{ path: string, info: {***REMOVED*** ***REMOVED******REMOVED***
   */
  replacePlaceholdersInFilename(compiler, filename, fileContent, compilation) {
    if (/\[\\*([\w:]+)\\*\]/i.test(filename) === false) {
      return { path: filename, info: {***REMOVED*** ***REMOVED***;
    ***REMOVED***

    const hash = compiler.webpack.util.createHash(
      compilation.outputOptions.hashFunction,
    );

    hash.update(fileContent);

    if (compilation.outputOptions.hashSalt) {
      hash.update(compilation.outputOptions.hashSalt);
    ***REMOVED***

    const contentHash = /** @type {string***REMOVED*** */ (
      hash
        .digest(compilation.outputOptions.hashDigest)
        .slice(0, compilation.outputOptions.hashDigestLength)
    );

    return compilation.getPathWithInfo(filename, {
      contentHash,
      chunk: {
        hash: contentHash,
        // @ts-ignore
        contentHash,
      ***REMOVED***,
    ***REMOVED***);
  ***REMOVED***

  /**
   * Function to generate HTML file.
   *
   * @private
   * @param {Compiler***REMOVED*** compiler
   * @param {Compilation***REMOVED*** compilation
   * @param {string***REMOVED*** outputName
   * @param {CachedChildCompilation***REMOVED*** childCompilerPlugin
   * @param {PreviousEmittedAssets***REMOVED*** previousEmittedAssets
   * @param {{ value: string | undefined ***REMOVED******REMOVED*** assetJson
   * @param {(err?: Error) => void***REMOVED*** callback
   */
  generateHTML(
    compiler,
    compilation,
    outputName,
    childCompilerPlugin,
    previousEmittedAssets,
    assetJson,
    callback,
  ) {
    // Get all entry point names for this html file
    const entryNames = Array.from(compilation.entrypoints.keys());
    const filteredEntryNames = this.filterEntryChunks(
      entryNames,
      this.options.chunks,
      this.options.excludeChunks,
    );
    const sortedEntryNames = this.sortEntryChunks(
      filteredEntryNames,
      this.options.chunksSortMode,
      compilation,
    );
    const templateResult = this.options.templateContent
      ? { mainCompilationHash: compilation.hash ***REMOVED***
      : childCompilerPlugin.getCompilationEntryResult(this.options.template);

    if ("error" in templateResult) {
      compilation.errors.push(
        prettyError(templateResult.error, compiler.context).toString(),
      );
    ***REMOVED***

    // If the child compilation was not executed during a previous main compile run
    // it is a cached result
    const isCompilationCached =
      templateResult.mainCompilationHash !== compilation.hash;
    /** Generated file paths from the entry point names */
    const assetsInformationByGroups = this.getAssetsInformationByGroups(
      compilation,
      outputName,
      sortedEntryNames,
    );
    // If the template and the assets did not change we don't have to emit the html
    const newAssetJson = JSON.stringify(
      this.getAssetFiles(assetsInformationByGroups),
    );

    if (
      isCompilationCached &&
      this.options.cache &&
      assetJson.value === newAssetJson
    ) {
      previousEmittedAssets.forEach(({ name, source, info ***REMOVED***) => {
        compilation.emitAsset(name, source, info);
      ***REMOVED***);
      return callback();
    ***REMOVED*** else {
      previousEmittedAssets.length = 0;
      assetJson.value = newAssetJson;
    ***REMOVED***

    // The html-webpack plugin uses a object representation for the html-tags which will be injected
    // to allow altering them more easily
    // Just before they are converted a third-party-plugin author might change the order and content
    const assetsPromise = this.generateFavicon(
      compiler,
      this.options.favicon,
      compilation,
      assetsInformationByGroups.publicPath,
      previousEmittedAssets,
    ).then((faviconPath) => {
      assetsInformationByGroups.favicon = faviconPath;
      return HtmlWebpackPlugin.getCompilationHooks(
        compilation,
      ).beforeAssetTagGeneration.promise({
        assets: assetsInformationByGroups,
        outputName,
        plugin: this,
      ***REMOVED***);
    ***REMOVED***);

    // Turn the js and css paths into grouped HtmlTagObjects
    const assetTagGroupsPromise = assetsPromise
      // And allow third-party-plugin authors to reorder and change the assetTags before they are grouped
      .then(({ assets ***REMOVED***) =>
        HtmlWebpackPlugin.getCompilationHooks(
          compilation,
        ).alterAssetTags.promise({
          assetTags: {
            scripts: this.generatedScriptTags(assets.js),
            styles: this.generateStyleTags(assets.css),
            meta: [
              ...(this.options.base !== false
                ? this.generateBaseTag(this.options.base)
                : []),
              ...this.generatedMetaTags(this.options.meta),
              ...(assets.favicon
                ? this.generateFaviconTag(assets.favicon)
                : []),
            ],
          ***REMOVED***,
          outputName,
          publicPath: assetsInformationByGroups.publicPath,
          plugin: this,
        ***REMOVED***),
      )
      .then(({ assetTags ***REMOVED***) => {
        // Inject scripts to body unless it set explicitly to head
        const scriptTarget =
          this.options.inject === "head" ||
          (this.options.inject !== "body" &&
            this.options.scriptLoading !== "blocking")
            ? "head"
            : "body";
        // Group assets to `head` and `body` tag arrays
        const assetGroups = this.groupAssetsByElements(assetTags, scriptTarget);
        // Allow third-party-plugin authors to reorder and change the assetTags once they are grouped
        return HtmlWebpackPlugin.getCompilationHooks(
          compilation,
        ).alterAssetTagGroups.promise({
          headTags: assetGroups.headTags,
          bodyTags: assetGroups.bodyTags,
          outputName,
          publicPath: assetsInformationByGroups.publicPath,
          plugin: this,
        ***REMOVED***);
      ***REMOVED***);

    // Turn the compiled template into a nodejs function or into a nodejs string
    const templateEvaluationPromise = Promise.resolve().then(() => {
      if ("error" in templateResult) {
        return this.options.showErrors
          ? prettyError(templateResult.error, compiler.context).toHtml()
          : "ERROR";
      ***REMOVED***

      // Allow to use a custom function / string instead
      if (this.options.templateContent !== false) {
        return this.options.templateContent;
      ***REMOVED***

      // Once everything is compiled evaluate the html factory and replace it with its content
      if ("compiledEntry" in templateResult) {
        const compiledEntry = templateResult.compiledEntry;
        const assets = compiledEntry.assets;

        // Store assets from child compiler to re-emit them later
        for (const name in assets) {
          previousEmittedAssets.push({
            name,
            source: assets[name].source,
            info: assets[name].info,
          ***REMOVED***);
        ***REMOVED***

        return this.evaluateCompilationResult(
          compiledEntry.content,
          assetsInformationByGroups.publicPath,
          this.options.template,
        );
      ***REMOVED***

      return Promise.reject(
        new Error("Child compilation contained no compiledEntry"),
      );
    ***REMOVED***);
    const templateExecutionPromise = Promise.all([
      assetsPromise,
      assetTagGroupsPromise,
      templateEvaluationPromise,
    ])
      // Execute the template
      .then(([assetsHookResult, assetTags, compilationResult]) =>
        typeof compilationResult !== "function"
          ? compilationResult
          : this.executeTemplate(
              compilationResult,
              assetsHookResult.assets,
              { headTags: assetTags.headTags, bodyTags: assetTags.bodyTags ***REMOVED***,
              compilation,
            ),
      );

    const injectedHtmlPromise = Promise.all([
      assetTagGroupsPromise,
      templateExecutionPromise,
    ])
      // Allow plugins to change the html before assets are injected
      .then(([assetTags, html]) => {
        const pluginArgs = {
          html,
          headTags: assetTags.headTags,
          bodyTags: assetTags.bodyTags,
          plugin: this,
          outputName,
        ***REMOVED***;
        return HtmlWebpackPlugin.getCompilationHooks(
          compilation,
        ).afterTemplateExecution.promise(pluginArgs);
      ***REMOVED***)
      .then(({ html, headTags, bodyTags ***REMOVED***) => {
        return this.postProcessHtml(compiler, html, assetsInformationByGroups, {
          headTags,
          bodyTags,
        ***REMOVED***);
      ***REMOVED***);

    const emitHtmlPromise = injectedHtmlPromise
      // Allow plugins to change the html after assets are injected
      .then((html) => {
        const pluginArgs = { html, plugin: this, outputName ***REMOVED***;
        return HtmlWebpackPlugin.getCompilationHooks(compilation)
          .beforeEmit.promise(pluginArgs)
          .then((result) => result.html);
      ***REMOVED***)
      .catch((err) => {
        // In case anything went wrong the promise is resolved
        // with the error message and an error is logged
        compilation.errors.push(prettyError(err, compiler.context).toString());
        return this.options.showErrors
          ? prettyError(err, compiler.context).toHtml()
          : "ERROR";
      ***REMOVED***)
      .then((html) => {
        const filename = outputName.replace(
          /\[templatehash([^\]]*)\]/g,
          require("util").deprecate(
            (match, options) => `[contenthash${options***REMOVED***]`,
            "[templatehash] is now [contenthash]",
          ),
        );
        const replacedFilename = this.replacePlaceholdersInFilename(
          compiler,
          filename,
          html,
          compilation,
        );
        const source = new compiler.webpack.sources.RawSource(html, false);

        // Add the evaluated html code to the webpack assets
        compilation.emitAsset(
          replacedFilename.path,
          source,
          replacedFilename.info,
        );
        previousEmittedAssets.push({ name: replacedFilename.path, source ***REMOVED***);

        return replacedFilename.path;
      ***REMOVED***)
      .then((finalOutputName) =>
        HtmlWebpackPlugin.getCompilationHooks(compilation)
          .afterEmit.promise({
            outputName: finalOutputName,
            plugin: this,
          ***REMOVED***)
          .catch((err) => {
            /** @type {Logger***REMOVED*** */
            (this.logger).error(err);
            return null;
          ***REMOVED***)
          .then(() => null),
      );

    // Once all files are added to the webpack compilation
    // let the webpack compiler continue
    emitHtmlPromise.then(() => {
      callback();
    ***REMOVED***);
  ***REMOVED***
***REMOVED***

/**
 * The default for options.templateParameter
 * Generate the template parameters
 *
 * Generate the template parameters for the template function
 * @param {Compilation***REMOVED*** compilation
 * @param {AssetsInformationByGroups***REMOVED*** assets
 * @param {{
     headTags: HtmlTagObject[],
     bodyTags: HtmlTagObject[]
   ***REMOVED******REMOVED*** assetTags
 * @param {ProcessedHtmlWebpackOptions***REMOVED*** options
 * @returns {TemplateParameter***REMOVED***
 */
function templateParametersGenerator(compilation, assets, assetTags, options) {
  return {
    compilation: compilation,
    webpackConfig: compilation.options,
    htmlWebpackPlugin: {
      tags: assetTags,
      files: assets,
      options: options,
    ***REMOVED***,
  ***REMOVED***;
***REMOVED***

// Statics:
/**
 * The major version number of this plugin
 */
HtmlWebpackPlugin.version = 5;

/**
 * A static helper to get the hooks for this plugin
 *
 * Usage: HtmlWebpackPlugin.getHooks(compilation).HOOK_NAME.tapAsync('YourPluginName', () => { ... ***REMOVED***);
 */
// TODO remove me in the next major release in favor getCompilationHooks
HtmlWebpackPlugin.getHooks = HtmlWebpackPlugin.getCompilationHooks;
HtmlWebpackPlugin.createHtmlTagObject = createHtmlTagObject;

module.exports = HtmlWebpackPlugin;
