// yaml/browser/dist/stringify/stringifyCollection.js
import { isNode, isPair ***REMOVED*** from '../nodes/identity.js';
import { stringify ***REMOVED*** from './stringify.js';
import { lineComment, indentComment ***REMOVED*** from './stringifyComment.js';

function stringifyCollection(collection, ctx, options) {
    const flow = ctx.inFlow ?? collection.flow;
    const stringify = flow ? stringifyFlowCollection : stringifyBlockCollection;
    return stringify(collection, ctx, options);
***REMOVED***
function stringifyBlockCollection({ comment, items ***REMOVED***, ctx, { blockItemPrefix, flowChars, itemIndent, onChompKeep, onComment ***REMOVED***) {
    const { indent, options: { commentString ***REMOVED*** ***REMOVED*** = ctx;
    const itemCtx = Object.assign({***REMOVED***, ctx, { indent: itemIndent, type: null ***REMOVED***);
    let chompKeep = false; // flag for the preceding node's status
    const lines = [];
    for (let i = 0; i < items.length; ++i) {
        const item = items[i];
        let comment = null;
        if (isNode(item)) {
            if (!chompKeep && item.spaceBefore)
                lines.push('');
            addCommentBefore(ctx, lines, item.commentBefore, chompKeep);
            if (item.comment)
                comment = item.comment;
        ***REMOVED***
        else if (isPair(item)) {
            const ik = isNode(item.key) ? item.key : null;
            if (ik) {
                if (!chompKeep && ik.spaceBefore)
                    lines.push('');
                addCommentBefore(ctx, lines, ik.commentBefore, chompKeep);
            ***REMOVED***
        ***REMOVED***
        chompKeep = false;
        let str = stringify(item, itemCtx, () => (comment = null), () => (chompKeep = true));
        if (comment)
            str += lineComment(str, itemIndent, commentString(comment));
        if (chompKeep && comment)
            chompKeep = false;
        lines.push(blockItemPrefix + str);
    ***REMOVED***
    let str;
    if (lines.length === 0) {
        str = flowChars.start + flowChars.end;
    ***REMOVED***
    else {
        str = lines[0];
        for (let i = 1; i < lines.length; ++i) {
            const line = lines[i];
            str += line ? `\n${indent***REMOVED***${line***REMOVED***` : '\n';
        ***REMOVED***
    ***REMOVED***
    if (comment) {
        str += '\n' + indentComment(commentString(comment), indent);
        if (onComment)
            onComment();
    ***REMOVED***
    else if (chompKeep && onChompKeep)
        onChompKeep();
    return str;
***REMOVED***
function stringifyFlowCollection({ items ***REMOVED***, ctx, { flowChars, itemIndent ***REMOVED***) {
    const { indent, indentStep, flowCollectionPadding: fcPadding, options: { commentString ***REMOVED*** ***REMOVED*** = ctx;
    itemIndent += indentStep;
    const itemCtx = Object.assign({***REMOVED***, ctx, {
        indent: itemIndent,
        inFlow: true,
        type: null
    ***REMOVED***);
    let reqNewline = false;
    let linesAtValue = 0;
    const lines = [];
    for (let i = 0; i < items.length; ++i) {
        const item = items[i];
        let comment = null;
        if (isNode(item)) {
            if (item.spaceBefore)
                lines.push('');
            addCommentBefore(ctx, lines, item.commentBefore, false);
            if (item.comment)
                comment = item.comment;
        ***REMOVED***
        else if (isPair(item)) {
            const ik = isNode(item.key) ? item.key : null;
            if (ik) {
                if (ik.spaceBefore)
                    lines.push('');
                addCommentBefore(ctx, lines, ik.commentBefore, false);
                if (ik.comment)
                    reqNewline = true;
            ***REMOVED***
            const iv = isNode(item.value) ? item.value : null;
            if (iv) {
                if (iv.comment)
                    comment = iv.comment;
                if (iv.commentBefore)
                    reqNewline = true;
            ***REMOVED***
            else if (item.value == null && ik?.comment) {
                comment = ik.comment;
            ***REMOVED***
        ***REMOVED***
        if (comment)
            reqNewline = true;
        let str = stringify(item, itemCtx, () => (comment = null));
        if (i < items.length - 1)
            str += ',';
        if (comment)
            str += lineComment(str, itemIndent, commentString(comment));
        if (!reqNewline && (lines.length > linesAtValue || str.includes('\n')))
            reqNewline = true;
        lines.push(str);
        linesAtValue = lines.length;
    ***REMOVED***
    const { start, end ***REMOVED*** = flowChars;
    if (lines.length === 0) {
        return start + end;
    ***REMOVED***
    else {
        if (!reqNewline) {
            const len = lines.reduce((sum, line) => sum + line.length + 2, 2);
            reqNewline = ctx.options.lineWidth > 0 && len > ctx.options.lineWidth;
        ***REMOVED***
        if (reqNewline) {
            let str = start;
            for (const line of lines)
                str += line ? `\n${indentStep***REMOVED***${indent***REMOVED***${line***REMOVED***` : '\n';
            return `${str***REMOVED***\n${indent***REMOVED***${end***REMOVED***`;
        ***REMOVED***
        else {
            return `${start***REMOVED***${fcPadding***REMOVED***${lines.join(' ')***REMOVED***${fcPadding***REMOVED***${end***REMOVED***`;
        ***REMOVED***
    ***REMOVED***
***REMOVED***
function addCommentBefore({ indent, options: { commentString ***REMOVED*** ***REMOVED***, lines, comment, chompKeep) {
    if (comment && chompKeep)
        comment = comment.replace(/^\n+/, '');
    if (comment) {
        const ic = indentComment(commentString(comment), indent);
        lines.push(ic.trimStart()); // Avoid double indent on first line
    ***REMOVED***
***REMOVED***

export { stringifyCollection ***REMOVED***;
