// ajv/dist/refs/jtd-schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const shared = (root) => {
    const sch = {
        nullable: { type: "boolean" ***REMOVED***,
        metadata: {
            optionalProperties: {
                union: { elements: { ref: "schema" ***REMOVED*** ***REMOVED***,
            ***REMOVED***,
            additionalProperties: true,
        ***REMOVED***,
    ***REMOVED***;
    if (root)
        sch.definitions = { values: { ref: "schema" ***REMOVED*** ***REMOVED***;
    return sch;
***REMOVED***;
const emptyForm = (root) => ({
    optionalProperties: shared(root),
***REMOVED***);
const refForm = (root) => ({
    properties: {
        ref: { type: "string" ***REMOVED***,
    ***REMOVED***,
    optionalProperties: shared(root),
***REMOVED***);
const typeForm = (root) => ({
    properties: {
        type: {
            enum: [
                "boolean",
                "timestamp",
                "string",
                "float32",
                "float64",
                "int8",
                "uint8",
                "int16",
                "uint16",
                "int32",
                "uint32",
            ],
        ***REMOVED***,
    ***REMOVED***,
    optionalProperties: shared(root),
***REMOVED***);
const enumForm = (root) => ({
    properties: {
        enum: { elements: { type: "string" ***REMOVED*** ***REMOVED***,
    ***REMOVED***,
    optionalProperties: shared(root),
***REMOVED***);
const elementsForm = (root) => ({
    properties: {
        elements: { ref: "schema" ***REMOVED***,
    ***REMOVED***,
    optionalProperties: shared(root),
***REMOVED***);
const propertiesForm = (root) => ({
    properties: {
        properties: { values: { ref: "schema" ***REMOVED*** ***REMOVED***,
    ***REMOVED***,
    optionalProperties: {
        optionalProperties: { values: { ref: "schema" ***REMOVED*** ***REMOVED***,
        additionalProperties: { type: "boolean" ***REMOVED***,
        ...shared(root),
    ***REMOVED***,
***REMOVED***);
const optionalPropertiesForm = (root) => ({
    properties: {
        optionalProperties: { values: { ref: "schema" ***REMOVED*** ***REMOVED***,
    ***REMOVED***,
    optionalProperties: {
        additionalProperties: { type: "boolean" ***REMOVED***,
        ...shared(root),
    ***REMOVED***,
***REMOVED***);
const discriminatorForm = (root) => ({
    properties: {
        discriminator: { type: "string" ***REMOVED***,
        mapping: {
            values: {
                metadata: {
                    union: [propertiesForm(false), optionalPropertiesForm(false)],
                ***REMOVED***,
            ***REMOVED***,
        ***REMOVED***,
    ***REMOVED***,
    optionalProperties: shared(root),
***REMOVED***);
const valuesForm = (root) => ({
    properties: {
        values: { ref: "schema" ***REMOVED***,
    ***REMOVED***,
    optionalProperties: shared(root),
***REMOVED***);
const schema = (root) => ({
    metadata: {
        union: [
            emptyForm,
            refForm,
            typeForm,
            enumForm,
            elementsForm,
            propertiesForm,
            optionalPropertiesForm,
            discriminatorForm,
            valuesForm,
        ].map((s) => s(root)),
    ***REMOVED***,
***REMOVED***);
const jtdMetaSchema = {
    definitions: {
        schema: schema(false),
    ***REMOVED***,
    ...schema(true),
***REMOVED***;
exports.default = jtdMetaSchema;
