// @babel/plugin-transform-react-constant-elements/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var _helperPluginUtils = require("@babel/helper-plugin-utils");
var _core = require("@babel/core");
var _default = exports.default = (0, _helperPluginUtils.declare)((api, options) => {
  api.assertVersion(7);
  const {
    allowMutablePropsOnTags
  ***REMOVED*** = options;
  if (allowMutablePropsOnTags != null && !Array.isArray(allowMutablePropsOnTags)) {
    throw new Error(".allowMutablePropsOnTags must be an array, null, or undefined.");
  ***REMOVED***
  const HOISTED = new WeakMap();
  function declares(node, scope) {
    if (_core.types.isJSXIdentifier(node, {
      name: "this"
    ***REMOVED***) || _core.types.isJSXIdentifier(node, {
      name: "arguments"
    ***REMOVED***) || _core.types.isJSXIdentifier(node, {
      name: "super"
    ***REMOVED***) || _core.types.isJSXIdentifier(node, {
      name: "new"
    ***REMOVED***)) {
      const {
        path
      ***REMOVED*** = scope;
      return path.isFunctionParent() && !path.isArrowFunctionExpression();
    ***REMOVED***
    return scope.hasOwnBinding(node.name);
  ***REMOVED***
  function isHoistingScope({
    path
  ***REMOVED***) {
    return path.isFunctionParent() || path.isLoop() || path.isProgram();
  ***REMOVED***
  function getHoistingScope(scope) {
    while (!isHoistingScope(scope)) scope = scope.parent;
    return scope;
  ***REMOVED***
  const targetScopeVisitor = {
    ReferencedIdentifier(path, state) {
      const {
        node
      ***REMOVED*** = path;
      let {
        scope
      ***REMOVED*** = path;
      while (scope !== state.jsxScope) {
        if (declares(node, scope)) return;
        scope = scope.parent;
      ***REMOVED***
      while (scope) {
        if (scope === state.targetScope) return;
        if (declares(node, scope)) break;
        scope = scope.parent;
      ***REMOVED***
      state.targetScope = getHoistingScope(scope);
    ***REMOVED***
  ***REMOVED***;
  const immutabilityVisitor = {
    enter(path, state) {
      var _expressionResult$deo;
      const stop = () => {
        state.isImmutable = false;
        path.stop();
      ***REMOVED***;
      const skip = () => {
        path.skip();
      ***REMOVED***;
      if (path.isJSXClosingElement()) {
        skip();
        return;
      ***REMOVED***
      if (path.isJSXIdentifier({
        name: "ref"
      ***REMOVED***) && path.parentPath.isJSXAttribute({
        name: path.node
      ***REMOVED***)) {
        stop();
        return;
      ***REMOVED***
      if (path.isJSXIdentifier() || path.isJSXMemberExpression() || path.isJSXNamespacedName() || path.isImmutable()) {
        return;
      ***REMOVED***
      if (path.isIdentifier()) {
        const binding = path.scope.getBinding(path.node.name);
        if (binding != null && binding.constant) return;
      ***REMOVED***
      const {
        mutablePropsAllowed
      ***REMOVED*** = state;
      if (mutablePropsAllowed && path.isFunction()) {
        path.traverse(targetScopeVisitor, state);
        skip();
        return;
      ***REMOVED***
      if (!path.isPure()) {
        stop();
        return;
      ***REMOVED***
      const expressionResult = path.evaluate();
      if (expressionResult.confident) {
        const {
          value
        ***REMOVED*** = expressionResult;
        if (mutablePropsAllowed || value === null || typeof value !== "object" && typeof value !== "function") {
          skip();
          return;
        ***REMOVED***
      ***REMOVED*** else if ((_expressionResult$deo = expressionResult.deopt) != null && _expressionResult$deo.isIdentifier()) {
        return;
      ***REMOVED***
      stop();
    ***REMOVED***
  ***REMOVED***;
  const hoistingVisitor = Object.assign({***REMOVED***, immutabilityVisitor, targetScopeVisitor);
  return {
    name: "transform-react-constant-elements",
    visitor: {
      "JSXElement|JSXFragment"(path) {
        if (HOISTED.has(path.node)) return;
        let mutablePropsAllowed = false;
        let name;
        if (path.isJSXElement()) {
          name = path.node.openingElement.name;
          if (allowMutablePropsOnTags != null) {
            let lastSegment = name;
            while (_core.types.isJSXMemberExpression(lastSegment)) {
              lastSegment = lastSegment.property;
            ***REMOVED***
            const elementName = lastSegment.name;
            mutablePropsAllowed = allowMutablePropsOnTags.includes(elementName);
          ***REMOVED***
        ***REMOVED*** else {
          name = path.node;
        ***REMOVED***
        let jsxScope;
        let current = path;
        while (!jsxScope && current.parentPath.isJSX()) {
          current = current.parentPath;
          jsxScope = HOISTED.get(current.node);
        ***REMOVED***
        jsxScope != null ? jsxScope : jsxScope = path.scope;
        HOISTED.set(path.node, jsxScope);
        const visitorState = {
          isImmutable: true,
          mutablePropsAllowed,
          jsxScope,
          targetScope: path.scope.getProgramParent()
        ***REMOVED***;
        path.traverse(hoistingVisitor, visitorState);
        if (!visitorState.isImmutable) return;
        const {
          targetScope
        ***REMOVED*** = visitorState;
        for (let currentScope = jsxScope;;) {
          if (targetScope === currentScope) return;
          if (isHoistingScope(currentScope)) break;
          currentScope = currentScope.parent;
          if (!currentScope) {
            throw new Error("Internal @babel/plugin-transform-react-constant-elements error: " + "targetScope must be an ancestor of jsxScope. " + "This is a Babel bug, please report it.");
          ***REMOVED***
        ***REMOVED***
        const id = path.scope.generateUidBasedOnNode(name);
        targetScope.push({
          id: _core.types.identifier(id)
        ***REMOVED***);
        HOISTED.set(path.node, targetScope);
        let replacement = _core.template.expression.ast`
          ${_core.types.identifier(id)***REMOVED*** || (${_core.types.identifier(id)***REMOVED*** = ${path.node***REMOVED***)
        `;
        if (path.parentPath.isJSXElement() || path.parentPath.isJSXAttribute() || path.parentPath.isJSXFragment()) {
          replacement = _core.types.jsxExpressionContainer(replacement);
        ***REMOVED***
        path.replaceWith(replacement);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;
***REMOVED***);
