// style-loader/dist/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.getExportLazyStyleCode = getExportLazyStyleCode;
exports.getExportStyleCode = getExportStyleCode;
exports.getImportInsertBySelectorCode = getImportInsertBySelectorCode;
exports.getImportInsertStyleElementCode = getImportInsertStyleElementCode;
exports.getImportIsOldIECode = getImportIsOldIECode;
exports.getImportLinkAPICode = getImportLinkAPICode;
exports.getImportLinkContentCode = getImportLinkContentCode;
exports.getImportStyleAPICode = getImportStyleAPICode;
exports.getImportStyleContentCode = getImportStyleContentCode;
exports.getImportStyleDomAPICode = getImportStyleDomAPICode;
exports.getInsertOptionCode = getInsertOptionCode;
exports.getLinkHmrCode = getLinkHmrCode;
exports.getSetAttributesCode = getSetAttributesCode;
exports.getStyleHmrCode = getStyleHmrCode;
exports.getStyleTagTransformFn = getStyleTagTransformFn;
exports.getStyleTagTransformFnCode = getStyleTagTransformFnCode;
exports.getdomAPI = getdomAPI;
exports.stringifyRequest = stringifyRequest;
var _path = _interopRequireDefault(require("path"));
var _isEqualLocals = _interopRequireDefault(require("./runtime/isEqualLocals"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj ***REMOVED***; ***REMOVED***
const matchRelativePath = /^\.\.?[/\\]/;
function isAbsolutePath(str) {
  return _path.default.posix.isAbsolute(str) || _path.default.win32.isAbsolute(str);
***REMOVED***
function isRelativePath(str) {
  return matchRelativePath.test(str);
***REMOVED***

// TODO simplify for the next major release
function stringifyRequest(loaderContext, request) {
  if (typeof loaderContext.utils !== "undefined" && typeof loaderContext.utils.contextify === "function") {
    return JSON.stringify(loaderContext.utils.contextify(loaderContext.context, request));
  ***REMOVED***
  const splitted = request.split("!");
  const {
    context
  ***REMOVED*** = loaderContext;
  return JSON.stringify(splitted.map(part => {
    // First, separate singlePath from query, because the query might contain paths again
    const splittedPart = part.match(/^(.*?)(\?.*)/);
    const query = splittedPart ? splittedPart[2] : "";
    let singlePath = splittedPart ? splittedPart[1] : part;
    if (isAbsolutePath(singlePath) && context) {
      singlePath = _path.default.relative(context, singlePath);
      if (isAbsolutePath(singlePath)) {
        // If singlePath still matches an absolute path, singlePath was on a different drive than context.
        // In this case, we leave the path platform-specific without replacing any separators.
        // @see https://github.com/webpack/loader-utils/pull/14
        return singlePath + query;
      ***REMOVED***
      if (isRelativePath(singlePath) === false) {
        // Ensure that the relative path starts at least with ./ otherwise it would be a request into the modules directory (like node_modules).
        singlePath = `./${singlePath***REMOVED***`;
      ***REMOVED***
    ***REMOVED***
    return singlePath.replace(/\\/g, "/") + query;
  ***REMOVED***).join("!"));
***REMOVED***
function getImportLinkAPICode(esModule, loaderContext) {
  const modulePath = stringifyRequest(loaderContext, `!${_path.default.join(__dirname, "runtime/injectStylesIntoLinkTag.js")***REMOVED***`);
  return esModule ? `import API from ${modulePath***REMOVED***;` : `var API = require(${modulePath***REMOVED***);`;
***REMOVED***
function getImportLinkContentCode(esModule, loaderContext, request) {
  const modulePath = stringifyRequest(loaderContext, `!!${request***REMOVED***`);
  return esModule ? `import content from ${modulePath***REMOVED***;` : `var content = require(${modulePath***REMOVED***);`;
***REMOVED***
function getImportStyleAPICode(esModule, loaderContext) {
  const modulePath = stringifyRequest(loaderContext, `!${_path.default.join(__dirname, "runtime/injectStylesIntoStyleTag.js")***REMOVED***`);
  return esModule ? `import API from ${modulePath***REMOVED***;` : `var API = require(${modulePath***REMOVED***);`;
***REMOVED***
function getImportStyleDomAPICode(esModule, loaderContext, isSingleton, isAuto) {
  const styleAPI = stringifyRequest(loaderContext, `!${_path.default.join(__dirname, "runtime/styleDomAPI.js")***REMOVED***`);
  const singletonAPI = stringifyRequest(loaderContext, `!${_path.default.join(__dirname, "runtime/singletonStyleDomAPI.js")***REMOVED***`);
  if (isAuto) {
    return esModule ? `import domAPI from ${styleAPI***REMOVED***;
        import domAPISingleton from ${singletonAPI***REMOVED***;` : `var domAPI = require(${styleAPI***REMOVED***);
        var domAPISingleton = require(${singletonAPI***REMOVED***);`;
  ***REMOVED***
  return esModule ? `import domAPI from ${isSingleton ? singletonAPI : styleAPI***REMOVED***;` : `var domAPI = require(${isSingleton ? singletonAPI : styleAPI***REMOVED***);`;
***REMOVED***
function getImportStyleContentCode(esModule, loaderContext, request) {
  const modulePath = stringifyRequest(loaderContext, `!!${request***REMOVED***`);
  return esModule ? `import content, * as namedExport from ${modulePath***REMOVED***;` : `var content = require(${modulePath***REMOVED***);`;
***REMOVED***
function getImportInsertBySelectorCode(esModule, loaderContext, insertType, options) {
  if (insertType === "selector") {
    const modulePath = stringifyRequest(loaderContext, `!${_path.default.join(__dirname, "runtime/insertBySelector.js")***REMOVED***`);
    return esModule ? `import insertFn from ${modulePath***REMOVED***;` : `var insertFn = require(${modulePath***REMOVED***);`;
  ***REMOVED***
  if (insertType === "module-path") {
    const modulePath = stringifyRequest(loaderContext, `${options.insert***REMOVED***`);
    loaderContext.addBuildDependency(options.insert);
    return esModule ? `import insertFn from ${modulePath***REMOVED***;` : `var insertFn = require(${modulePath***REMOVED***);`;
  ***REMOVED***
  return "";
***REMOVED***
function getInsertOptionCode(insertType, options) {
  if (insertType === "selector") {
    const insert = options.insert ? JSON.stringify(options.insert) : '"head"';
    return `
      options.insert = insertFn.bind(null, ${insert***REMOVED***);
    `;
  ***REMOVED***
  if (insertType === "module-path") {
    return `options.insert = insertFn;`;
  ***REMOVED***

  // Todo remove "function" type for insert option in next major release, because code duplication occurs. Leave require.resolve()
  return `options.insert = ${options.insert.toString()***REMOVED***;`;
***REMOVED***
function getImportInsertStyleElementCode(esModule, loaderContext) {
  const modulePath = stringifyRequest(loaderContext, `!${_path.default.join(__dirname, "runtime/insertStyleElement.js")***REMOVED***`);
  return esModule ? `import insertStyleElement from ${modulePath***REMOVED***;` : `var insertStyleElement = require(${modulePath***REMOVED***);`;
***REMOVED***
function getStyleHmrCode(esModule, loaderContext, request, lazy) {
  const modulePath = stringifyRequest(loaderContext, `!!${request***REMOVED***`);
  return `
if (module.hot) {
  if (!content.locals || module.hot.invalidate) {
    var isEqualLocals = ${_isEqualLocals.default.toString()***REMOVED***;
    var isNamedExport = ${esModule ? "!content.locals" : false***REMOVED***;
    var oldLocals = isNamedExport ? namedExport : content.locals;

    module.hot.accept(
      ${modulePath***REMOVED***,
      function () {
        ${esModule ? `if (!isEqualLocals(oldLocals, isNamedExport ? namedExport : content.locals, isNamedExport)) {
                module.hot.invalidate();

                return;
              ***REMOVED***

              oldLocals = isNamedExport ? namedExport : content.locals;

              ${lazy ? `if (update && refs > 0) {
                      update(content);
                    ***REMOVED***` : `update(content);`***REMOVED***` : `content = require(${modulePath***REMOVED***);

              content = content.__esModule ? content.default : content;

              ${lazy ? "" : `if (typeof content === 'string') {
                      content = [[module.id, content, '']];
                    ***REMOVED***`***REMOVED***

              if (!isEqualLocals(oldLocals, content.locals)) {
                module.hot.invalidate();

                return;
              ***REMOVED***

              oldLocals = content.locals;

              ${lazy ? `if (update && refs > 0) {
                        update(content);
                      ***REMOVED***` : `update(content);`***REMOVED***`***REMOVED***
      ***REMOVED***
    )
  ***REMOVED***

  module.hot.dispose(function() {
    ${lazy ? `if (update) {
            update();
          ***REMOVED***` : `update();`***REMOVED***
  ***REMOVED***);
***REMOVED***
`;
***REMOVED***
function getLinkHmrCode(esModule, loaderContext, request) {
  const modulePath = stringifyRequest(loaderContext, `!!${request***REMOVED***`);
  return `
if (module.hot) {
  module.hot.accept(
    ${modulePath***REMOVED***,
    function() {
     ${esModule ? "update(content);" : `content = require(${modulePath***REMOVED***);

           content = content.__esModule ? content.default : content;

           update(content);`***REMOVED***
    ***REMOVED***
  );

  module.hot.dispose(function() {
    update();
  ***REMOVED***);
***REMOVED***`;
***REMOVED***
function getdomAPI(isAuto) {
  return isAuto ? "isOldIE() ? domAPISingleton : domAPI" : "domAPI";
***REMOVED***
function getImportIsOldIECode(esModule, loaderContext) {
  const modulePath = stringifyRequest(loaderContext, `!${_path.default.join(__dirname, "runtime/isOldIE.js")***REMOVED***`);
  return esModule ? `import isOldIE from ${modulePath***REMOVED***;` : `var isOldIE = require(${modulePath***REMOVED***);`;
***REMOVED***
function getStyleTagTransformFnCode(esModule, loaderContext, options, isSingleton, styleTagTransformType) {
  if (isSingleton) {
    return "";
  ***REMOVED***
  if (styleTagTransformType === "default") {
    const modulePath = stringifyRequest(loaderContext, `!${_path.default.join(__dirname, "runtime/styleTagTransform.js")***REMOVED***`);
    return esModule ? `import styleTagTransformFn from ${modulePath***REMOVED***;` : `var styleTagTransformFn = require(${modulePath***REMOVED***);`;
  ***REMOVED***
  if (styleTagTransformType === "module-path") {
    const modulePath = stringifyRequest(loaderContext, `${options.styleTagTransform***REMOVED***`);
    loaderContext.addBuildDependency(options.styleTagTransform);
    return esModule ? `import styleTagTransformFn from ${modulePath***REMOVED***;` : `var styleTagTransformFn = require(${modulePath***REMOVED***);`;
  ***REMOVED***
  return "";
***REMOVED***
function getStyleTagTransformFn(options, isSingleton) {
  // Todo remove "function" type for styleTagTransform option in next major release, because code duplication occurs. Leave require.resolve()
  return isSingleton ? "" : typeof options.styleTagTransform === "function" ? `options.styleTagTransform = ${options.styleTagTransform.toString()***REMOVED***` : `options.styleTagTransform = styleTagTransformFn`;
***REMOVED***
function getExportStyleCode(esModule, loaderContext, request) {
  const modulePath = stringifyRequest(loaderContext, `!!${request***REMOVED***`);
  return esModule ? `export * from ${modulePath***REMOVED***;
       export default content && content.locals ? content.locals : undefined;` : "module.exports = content && content.locals || {***REMOVED***;";
***REMOVED***
function getExportLazyStyleCode(esModule, loaderContext, request) {
  const modulePath = stringifyRequest(loaderContext, `!!${request***REMOVED***`);
  return esModule ? `export * from ${modulePath***REMOVED***;
       export default exported;` : "module.exports = exported;";
***REMOVED***
function getSetAttributesCode(esModule, loaderContext, options) {
  let modulePath;
  if (typeof options.attributes !== "undefined") {
    modulePath = options.attributes.nonce !== "undefined" ? stringifyRequest(loaderContext, `!${_path.default.join(__dirname, "runtime/setAttributesWithAttributesAndNonce.js")***REMOVED***`) : stringifyRequest(loaderContext, `!${_path.default.join(__dirname, "runtime/setAttributesWithAttributes.js")***REMOVED***`);
  ***REMOVED*** else {
    modulePath = stringifyRequest(loaderContext, `!${_path.default.join(__dirname, "runtime/setAttributesWithoutAttributes.js")***REMOVED***`);
  ***REMOVED***
  return esModule ? `import setAttributes from ${modulePath***REMOVED***;` : `var setAttributes = require(${modulePath***REMOVED***);`;
***REMOVED***

// eslint-disable-next-line import/prefer-default-export
