// @rollup/pluginutils/dist/cjs/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true ***REMOVED***);

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; ***REMOVED***

var path = require('path');
var pm = _interopDefault(require('picomatch'));

const addExtension = function addExtension(filename, ext = '.js') {
    let result = `${filename***REMOVED***`;
    if (!path.extname(filename))
        result += ext;
    return result;
***REMOVED***;

function walk(ast, { enter, leave ***REMOVED***) {
	return visit(ast, null, enter, leave);
***REMOVED***

let should_skip = false;
let should_remove = false;
let replacement = null;
const context = {
	skip: () => should_skip = true,
	remove: () => should_remove = true,
	replace: (node) => replacement = node
***REMOVED***;

function replace(parent, prop, index, node) {
	if (parent) {
		if (index !== null) {
			parent[prop][index] = node;
		***REMOVED*** else {
			parent[prop] = node;
		***REMOVED***
	***REMOVED***
***REMOVED***

function remove(parent, prop, index) {
	if (parent) {
		if (index !== null) {
			parent[prop].splice(index, 1);
		***REMOVED*** else {
			delete parent[prop];
		***REMOVED***
	***REMOVED***
***REMOVED***

function visit(
	node,
	parent,
	enter,
	leave,
	prop,
	index
) {
	if (node) {
		if (enter) {
			const _should_skip = should_skip;
			const _should_remove = should_remove;
			const _replacement = replacement;
			should_skip = false;
			should_remove = false;
			replacement = null;

			enter.call(context, node, parent, prop, index);

			if (replacement) {
				node = replacement;
				replace(parent, prop, index, node);
			***REMOVED***

			if (should_remove) {
				remove(parent, prop, index);
			***REMOVED***

			const skipped = should_skip;
			const removed = should_remove;

			should_skip = _should_skip;
			should_remove = _should_remove;
			replacement = _replacement;

			if (skipped) return node;
			if (removed) return null;
		***REMOVED***

		for (const key in node) {
			const value = (node )[key];

			if (typeof value !== 'object') {
				continue;
			***REMOVED***

			else if (Array.isArray(value)) {
				for (let j = 0, k = 0; j < value.length; j += 1, k += 1) {
					if (value[j] !== null && typeof value[j].type === 'string') {
						if (!visit(value[j], node, enter, leave, key, k)) {
							// removed
							j--;
						***REMOVED***
					***REMOVED***
				***REMOVED***
			***REMOVED***

			else if (value !== null && typeof value.type === 'string') {
				visit(value, node, enter, leave, key, null);
			***REMOVED***
		***REMOVED***

		if (leave) {
			const _replacement = replacement;
			const _should_remove = should_remove;
			replacement = null;
			should_remove = false;

			leave.call(context, node, parent, prop, index);

			if (replacement) {
				node = replacement;
				replace(parent, prop, index, node);
			***REMOVED***

			if (should_remove) {
				remove(parent, prop, index);
			***REMOVED***

			const removed = should_remove;

			replacement = _replacement;
			should_remove = _should_remove;

			if (removed) return null;
		***REMOVED***
	***REMOVED***

	return node;
***REMOVED***

const extractors = {
    ArrayPattern(names, param) {
        for (const element of param.elements) {
            if (element)
                extractors[element.type](names, element);
        ***REMOVED***
    ***REMOVED***,
    AssignmentPattern(names, param) {
        extractors[param.left.type](names, param.left);
    ***REMOVED***,
    Identifier(names, param) {
        names.push(param.name);
    ***REMOVED***,
    MemberExpression() { ***REMOVED***,
    ObjectPattern(names, param) {
        for (const prop of param.properties) {
            // @ts-ignore Typescript reports that this is not a valid type
            if (prop.type === 'RestElement') {
                extractors.RestElement(names, prop);
            ***REMOVED***
            else {
                extractors[prop.value.type](names, prop.value);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***,
    RestElement(names, param) {
        extractors[param.argument.type](names, param.argument);
    ***REMOVED***
***REMOVED***;
const extractAssignedNames = function extractAssignedNames(param) {
    const names = [];
    extractors[param.type](names, param);
    return names;
***REMOVED***;

const blockDeclarations = {
    const: true,
    let: true
***REMOVED***;
class Scope {
    constructor(options = {***REMOVED***) {
        this.parent = options.parent;
        this.isBlockScope = !!options.block;
        this.declarations = Object.create(null);
        if (options.params) {
            options.params.forEach((param) => {
                extractAssignedNames(param).forEach((name) => {
                    this.declarations[name] = true;
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***
    addDeclaration(node, isBlockDeclaration, isVar) {
        if (!isBlockDeclaration && this.isBlockScope) {
            // it's a `var` or function node, and this
            // is a block scope, so we need to go up
            this.parent.addDeclaration(node, isBlockDeclaration, isVar);
        ***REMOVED***
        else if (node.id) {
            extractAssignedNames(node.id).forEach((name) => {
                this.declarations[name] = true;
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***
    contains(name) {
        return this.declarations[name] || (this.parent ? this.parent.contains(name) : false);
    ***REMOVED***
***REMOVED***
const attachScopes = function attachScopes(ast, propertyName = 'scope') {
    let scope = new Scope();
    walk(ast, {
        enter(n, parent) {
            const node = n;
            // function foo () {...***REMOVED***
            // class Foo {...***REMOVED***
            if (/(Function|Class)Declaration/.test(node.type)) {
                scope.addDeclaration(node, false, false);
            ***REMOVED***
            // var foo = 1
            if (node.type === 'VariableDeclaration') {
                const { kind ***REMOVED*** = node;
                const isBlockDeclaration = blockDeclarations[kind];
                // don't add const/let declarations in the body of a for loop #113
                const parentType = parent ? parent.type : '';
                if (!(isBlockDeclaration && /ForOfStatement/.test(parentType))) {
                    node.declarations.forEach((declaration) => {
                        scope.addDeclaration(declaration, isBlockDeclaration, true);
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
            let newScope;
            // create new function scope
            if (/Function/.test(node.type)) {
                const func = node;
                newScope = new Scope({
                    parent: scope,
                    block: false,
                    params: func.params
                ***REMOVED***);
                // named function expressions - the name is considered
                // part of the function's scope
                if (func.type === 'FunctionExpression' && func.id) {
                    newScope.addDeclaration(func, false, false);
                ***REMOVED***
            ***REMOVED***
            // create new block scope
            if (node.type === 'BlockStatement' && !/Function/.test(parent.type)) {
                newScope = new Scope({
                    parent: scope,
                    block: true
                ***REMOVED***);
            ***REMOVED***
            // catch clause has its own block scope
            if (node.type === 'CatchClause') {
                newScope = new Scope({
                    parent: scope,
                    params: node.param ? [node.param] : [],
                    block: true
                ***REMOVED***);
            ***REMOVED***
            if (newScope) {
                Object.defineProperty(node, propertyName, {
                    value: newScope,
                    configurable: true
                ***REMOVED***);
                scope = newScope;
            ***REMOVED***
        ***REMOVED***,
        leave(n) {
            const node = n;
            if (node[propertyName])
                scope = scope.parent;
        ***REMOVED***
    ***REMOVED***);
    return scope;
***REMOVED***;

// Helper since Typescript can't detect readonly arrays with Array.isArray
function isArray(arg) {
    return Array.isArray(arg);
***REMOVED***
function ensureArray(thing) {
    if (isArray(thing))
        return thing;
    if (thing == null)
        return [];
    return [thing];
***REMOVED***

function getMatcherString(id, resolutionBase) {
    if (resolutionBase === false) {
        return id;
    ***REMOVED***
    // resolve('') is valid and will default to process.cwd()
    const basePath = path.resolve(resolutionBase || '')
        .split(path.sep)
        .join('/')
        // escape all possible (posix + win) path characters that might interfere with regex
        .replace(/[-^$*+?.()|[\]{***REMOVED***]/g, '\\$&');
    // Note that we use posix.join because:
    // 1. the basePath has been normalized to use /
    // 2. the incoming glob (id) matcher, also uses /
    // otherwise Node will force backslash (\) on windows
    return path.posix.join(basePath, id);
***REMOVED***
const createFilter = function createFilter(include, exclude, options) {
    const resolutionBase = options && options.resolve;
    const getMatcher = (id) => id instanceof RegExp
        ? id
        : {
            test: (what) => {
                // this refactor is a tad overly verbose but makes for easy debugging
                const pattern = getMatcherString(id, resolutionBase);
                const fn = pm(pattern, { dot: true ***REMOVED***);
                const result = fn(what);
                return result;
            ***REMOVED***
        ***REMOVED***;
    const includeMatchers = ensureArray(include).map(getMatcher);
    const excludeMatchers = ensureArray(exclude).map(getMatcher);
    return function result(id) {
        if (typeof id !== 'string')
            return false;
        if (/\0/.test(id))
            return false;
        const pathId = id.split(path.sep).join('/');
        for (let i = 0; i < excludeMatchers.length; ++i) {
            const matcher = excludeMatchers[i];
            if (matcher.test(pathId))
                return false;
        ***REMOVED***
        for (let i = 0; i < includeMatchers.length; ++i) {
            const matcher = includeMatchers[i];
            if (matcher.test(pathId))
                return true;
        ***REMOVED***
        return !includeMatchers.length;
    ***REMOVED***;
***REMOVED***;

const reservedWords = 'break case class catch const continue debugger default delete do else export extends finally for function if import in instanceof let new return super switch this throw try typeof var void while with yield enum await implements package protected static interface private public';
const builtins = 'arguments Infinity NaN undefined null true false eval uneval isFinite isNaN parseFloat parseInt decodeURI decodeURIComponent encodeURI encodeURIComponent escape unescape Object Function Boolean Symbol Error EvalError InternalError RangeError ReferenceError SyntaxError TypeError URIError Number Math Date String RegExp Array Int8Array Uint8Array Uint8ClampedArray Int16Array Uint16Array Int32Array Uint32Array Float32Array Float64Array Map Set WeakMap WeakSet SIMD ArrayBuffer DataView JSON Promise Generator GeneratorFunction Reflect Proxy Intl';
const forbiddenIdentifiers = new Set(`${reservedWords***REMOVED*** ${builtins***REMOVED***`.split(' '));
forbiddenIdentifiers.add('');
const makeLegalIdentifier = function makeLegalIdentifier(str) {
    let identifier = str
        .replace(/-(\w)/g, (_, letter) => letter.toUpperCase())
        .replace(/[^$_a-zA-Z0-9]/g, '_');
    if (/\d/.test(identifier[0]) || forbiddenIdentifiers.has(identifier)) {
        identifier = `_${identifier***REMOVED***`;
    ***REMOVED***
    return identifier || '_';
***REMOVED***;

function stringify(obj) {
    return (JSON.stringify(obj) || 'undefined').replace(/[\u2028\u2029]/g, (char) => `\\u${`000${char.charCodeAt(0).toString(16)***REMOVED***`.slice(-4)***REMOVED***`);
***REMOVED***
function serializeArray(arr, indent, baseIndent) {
    let output = '[';
    const separator = indent ? `\n${baseIndent***REMOVED***${indent***REMOVED***` : '';
    for (let i = 0; i < arr.length; i++) {
        const key = arr[i];
        output += `${i > 0 ? ',' : ''***REMOVED***${separator***REMOVED***${serialize(key, indent, baseIndent + indent)***REMOVED***`;
    ***REMOVED***
    return `${output***REMOVED***${indent ? `\n${baseIndent***REMOVED***` : ''***REMOVED***]`;
***REMOVED***
function serializeObject(obj, indent, baseIndent) {
    let output = '{';
    const separator = indent ? `\n${baseIndent***REMOVED***${indent***REMOVED***` : '';
    const entries = Object.entries(obj);
    for (let i = 0; i < entries.length; i++) {
        const [key, value] = entries[i];
        const stringKey = makeLegalIdentifier(key) === key ? key : stringify(key);
        output += `${i > 0 ? ',' : ''***REMOVED***${separator***REMOVED***${stringKey***REMOVED***:${indent ? ' ' : ''***REMOVED***${serialize(value, indent, baseIndent + indent)***REMOVED***`;
    ***REMOVED***
    return `${output***REMOVED***${indent ? `\n${baseIndent***REMOVED***` : ''***REMOVED******REMOVED***`;
***REMOVED***
function serialize(obj, indent, baseIndent) {
    if (obj === Infinity)
        return 'Infinity';
    if (obj === -Infinity)
        return '-Infinity';
    if (obj === 0 && 1 / obj === -Infinity)
        return '-0';
    if (obj instanceof Date)
        return `new Date(${obj.getTime()***REMOVED***)`;
    if (obj instanceof RegExp)
        return obj.toString();
    if (obj !== obj)
        return 'NaN'; // eslint-disable-line no-self-compare
    if (Array.isArray(obj))
        return serializeArray(obj, indent, baseIndent);
    if (obj === null)
        return 'null';
    if (typeof obj === 'object')
        return serializeObject(obj, indent, baseIndent);
    return stringify(obj);
***REMOVED***
const dataToEsm = function dataToEsm(data, options = {***REMOVED***) {
    const t = options.compact ? '' : 'indent' in options ? options.indent : '\t';
    const _ = options.compact ? '' : ' ';
    const n = options.compact ? '' : '\n';
    const declarationType = options.preferConst ? 'const' : 'var';
    if (options.namedExports === false ||
        typeof data !== 'object' ||
        Array.isArray(data) ||
        data instanceof Date ||
        data instanceof RegExp ||
        data === null) {
        const code = serialize(data, options.compact ? null : t, '');
        const magic = _ || (/^[{[\-\/]/.test(code) ? '' : ' '); // eslint-disable-line no-useless-escape
        return `export default${magic***REMOVED***${code***REMOVED***;`;
    ***REMOVED***
    let namedExportCode = '';
    const defaultExportRows = [];
    for (const [key, value] of Object.entries(data)) {
        if (key === makeLegalIdentifier(key)) {
            if (options.objectShorthand)
                defaultExportRows.push(key);
            else
                defaultExportRows.push(`${key***REMOVED***:${_***REMOVED***${key***REMOVED***`);
            namedExportCode += `export ${declarationType***REMOVED*** ${key***REMOVED***${_***REMOVED***=${_***REMOVED***${serialize(value, options.compact ? null : t, '')***REMOVED***;${n***REMOVED***`;
        ***REMOVED***
        else {
            defaultExportRows.push(`${stringify(key)***REMOVED***:${_***REMOVED***${serialize(value, options.compact ? null : t, '')***REMOVED***`);
        ***REMOVED***
    ***REMOVED***
    return `${namedExportCode***REMOVED***export default${_***REMOVED***{${n***REMOVED***${t***REMOVED***${defaultExportRows.join(`,${n***REMOVED***${t***REMOVED***`)***REMOVED***${n***REMOVED******REMOVED***;${n***REMOVED***`;
***REMOVED***;

// TODO: remove this in next major
var index = {
    addExtension,
    attachScopes,
    createFilter,
    dataToEsm,
    extractAssignedNames,
    makeLegalIdentifier
***REMOVED***;

exports.addExtension = addExtension;
exports.attachScopes = attachScopes;
exports.createFilter = createFilter;
exports.dataToEsm = dataToEsm;
exports.default = index;
exports.extractAssignedNames = extractAssignedNames;
exports.makeLegalIdentifier = makeLegalIdentifier;
