// @babel/helper-define-polyfill-provider/lib/utils.js
"use strict";

exports.__esModule = true;
exports.createUtilsGetter = createUtilsGetter;
exports.getImportSource = getImportSource;
exports.getRequireSource = getRequireSource;
exports.has = has;
exports.intersection = intersection;
exports.resolveKey = resolveKey;
exports.resolveSource = resolveSource;
var _babel = _interopRequireWildcard(require("@babel/core"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; ***REMOVED***)(nodeInterop); ***REMOVED***
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; ***REMOVED*** if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj ***REMOVED***; ***REMOVED*** var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); ***REMOVED*** var newObj = {***REMOVED***; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); ***REMOVED*** else { newObj[key] = obj[key]; ***REMOVED*** ***REMOVED*** ***REMOVED*** newObj.default = obj; if (cache) { cache.set(obj, newObj); ***REMOVED*** return newObj; ***REMOVED***
const {
  types: t,
  template: template
***REMOVED*** = _babel.default || _babel;
function intersection(a, b) {
  const result = new Set();
  a.forEach(v => b.has(v) && result.add(v));
  return result;
***REMOVED***
function has(object, key) {
  return Object.prototype.hasOwnProperty.call(object, key);
***REMOVED***
function resolve(path, resolved = new Set()) {
  if (resolved.has(path)) return;
  resolved.add(path);
  if (path.isVariableDeclarator()) {
    if (path.get("id").isIdentifier()) {
      return resolve(path.get("init"), resolved);
    ***REMOVED***
  ***REMOVED*** else if (path.isReferencedIdentifier()) {
    const binding = path.scope.getBinding(path.node.name);
    if (!binding) return path;
    if (!binding.constant) return;
    return resolve(binding.path, resolved);
  ***REMOVED***
  return path;
***REMOVED***
function resolveId(path) {
  if (path.isIdentifier() && !path.scope.hasBinding(path.node.name, /* noGlobals */true)) {
    return path.node.name;
  ***REMOVED***
  const resolved = resolve(path);
  if (resolved != null && resolved.isIdentifier()) {
    return resolved.node.name;
  ***REMOVED***
***REMOVED***
function resolveKey(path, computed = false) {
  const {
    scope
  ***REMOVED*** = path;
  if (path.isStringLiteral()) return path.node.value;
  const isIdentifier = path.isIdentifier();
  if (isIdentifier && !(computed || path.parent.computed)) {
    return path.node.name;
  ***REMOVED***
  if (computed && path.isMemberExpression() && path.get("object").isIdentifier({
    name: "Symbol"
  ***REMOVED***) && !scope.hasBinding("Symbol", /* noGlobals */true)) {
    const sym = resolveKey(path.get("property"), path.node.computed);
    if (sym) return "Symbol." + sym;
  ***REMOVED***
  if (isIdentifier ? scope.hasBinding(path.node.name, /* noGlobals */true) : path.isPure()) {
    const {
      value
    ***REMOVED*** = path.evaluate();
    if (typeof value === "string") return value;
  ***REMOVED***
***REMOVED***
function resolveSource(obj) {
  if (obj.isMemberExpression() && obj.get("property").isIdentifier({
    name: "prototype"
  ***REMOVED***)) {
    const id = resolveId(obj.get("object"));
    if (id) {
      return {
        id,
        placement: "prototype"
      ***REMOVED***;
    ***REMOVED***
    return {
      id: null,
      placement: null
    ***REMOVED***;
  ***REMOVED***
  const id = resolveId(obj);
  if (id) {
    return {
      id,
      placement: "static"
    ***REMOVED***;
  ***REMOVED***
  const path = resolve(obj);
  switch (path == null ? void 0 : path.type) {
    case "RegExpLiteral":
      return {
        id: "RegExp",
        placement: "prototype"
      ***REMOVED***;
    case "FunctionExpression":
      return {
        id: "Function",
        placement: "prototype"
      ***REMOVED***;
    case "StringLiteral":
      return {
        id: "String",
        placement: "prototype"
      ***REMOVED***;
    case "NumberLiteral":
      return {
        id: "Number",
        placement: "prototype"
      ***REMOVED***;
    case "BooleanLiteral":
      return {
        id: "Boolean",
        placement: "prototype"
      ***REMOVED***;
    case "ObjectExpression":
      return {
        id: "Object",
        placement: "prototype"
      ***REMOVED***;
    case "ArrayExpression":
      return {
        id: "Array",
        placement: "prototype"
      ***REMOVED***;
  ***REMOVED***
  return {
    id: null,
    placement: null
  ***REMOVED***;
***REMOVED***
function getImportSource({
  node
***REMOVED***) {
  if (node.specifiers.length === 0) return node.source.value;
***REMOVED***
function getRequireSource({
  node
***REMOVED***) {
  if (!t.isExpressionStatement(node)) return;
  const {
    expression
  ***REMOVED*** = node;
  if (t.isCallExpression(expression) && t.isIdentifier(expression.callee) && expression.callee.name === "require" && expression.arguments.length === 1 && t.isStringLiteral(expression.arguments[0])) {
    return expression.arguments[0].value;
  ***REMOVED***
***REMOVED***
function hoist(node) {
  // @ts-expect-error
  node._blockHoist = 3;
  return node;
***REMOVED***
function createUtilsGetter(cache) {
  return path => {
    const prog = path.findParent(p => p.isProgram());
    return {
      injectGlobalImport(url, moduleName) {
        cache.storeAnonymous(prog, url, moduleName, (isScript, source) => {
          return isScript ? template.statement.ast`require(${source***REMOVED***)` : t.importDeclaration([], source);
        ***REMOVED***);
      ***REMOVED***,
      injectNamedImport(url, name, hint = name, moduleName) {
        return cache.storeNamed(prog, url, name, moduleName, (isScript, source, name) => {
          const id = prog.scope.generateUidIdentifier(hint);
          return {
            node: isScript ? hoist(template.statement.ast`
                  var ${id***REMOVED*** = require(${source***REMOVED***).${name***REMOVED***
                `) : t.importDeclaration([t.importSpecifier(id, name)], source),
            name: id.name
          ***REMOVED***;
        ***REMOVED***);
      ***REMOVED***,
      injectDefaultImport(url, hint = url, moduleName) {
        return cache.storeNamed(prog, url, "default", moduleName, (isScript, source) => {
          const id = prog.scope.generateUidIdentifier(hint);
          return {
            node: isScript ? hoist(template.statement.ast`var ${id***REMOVED*** = require(${source***REMOVED***)`) : t.importDeclaration([t.importDefaultSpecifier(id)], source),
            name: id.name
          ***REMOVED***;
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***;
***REMOVED***
