// @babel/plugin-transform-template-literals/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var _helperPluginUtils = require("@babel/helper-plugin-utils");
var _core = require("@babel/core");
var _default = exports.default = (0, _helperPluginUtils.declare)((api, options) => {
  var _api$assumption, _api$assumption2;
  api.assertVersion(7);
  const ignoreToPrimitiveHint = (_api$assumption = api.assumption("ignoreToPrimitiveHint")) != null ? _api$assumption : options.loose;
  const mutableTemplateObject = (_api$assumption2 = api.assumption("mutableTemplateObject")) != null ? _api$assumption2 : options.loose;
  let helperName = "taggedTemplateLiteral";
  if (mutableTemplateObject) helperName += "Loose";
  function buildConcatCallExpressions(items) {
    let avail = true;
    return items.reduce(function (left, right) {
      let canBeInserted = _core.types.isLiteral(right);
      if (!canBeInserted && avail) {
        canBeInserted = true;
        avail = false;
      ***REMOVED***
      if (canBeInserted && _core.types.isCallExpression(left)) {
        left.arguments.push(right);
        return left;
      ***REMOVED***
      return _core.types.callExpression(_core.types.memberExpression(left, _core.types.identifier("concat")), [right]);
    ***REMOVED***);
  ***REMOVED***
  return {
    name: "transform-template-literals",
    visitor: {
      TaggedTemplateExpression(path) {
        const {
          node
        ***REMOVED*** = path;
        const {
          quasi
        ***REMOVED*** = node;
        const strings = [];
        const raws = [];
        let isStringsRawEqual = true;
        for (const elem of quasi.quasis) {
          const {
            raw,
            cooked
          ***REMOVED*** = elem.value;
          const value = cooked == null ? path.scope.buildUndefinedNode() : _core.types.stringLiteral(cooked);
          strings.push(value);
          raws.push(_core.types.stringLiteral(raw));
          if (raw !== cooked) {
            isStringsRawEqual = false;
          ***REMOVED***
        ***REMOVED***
        const helperArgs = [_core.types.arrayExpression(strings)];
        if (!isStringsRawEqual) {
          helperArgs.push(_core.types.arrayExpression(raws));
        ***REMOVED***
        const tmp = path.scope.generateUidIdentifier("templateObject");
        path.scope.getProgramParent().push({
          id: _core.types.cloneNode(tmp)
        ***REMOVED***);
        path.replaceWith(_core.types.callExpression(node.tag, [_core.template.expression.ast`
              ${_core.types.cloneNode(tmp)***REMOVED*** || (
                ${tmp***REMOVED*** = ${this.addHelper(helperName)***REMOVED***(${helperArgs***REMOVED***)
              )
            `, ...quasi.expressions]));
      ***REMOVED***,
      TemplateLiteral(path) {
        if (path.parent.type === "TSLiteralType") {
          return;
        ***REMOVED***
        const nodes = [];
        const expressions = path.get("expressions");
        let index = 0;
        for (const elem of path.node.quasis) {
          if (elem.value.cooked) {
            nodes.push(_core.types.stringLiteral(elem.value.cooked));
          ***REMOVED***
          if (index < expressions.length) {
            const expr = expressions[index++];
            const node = expr.node;
            if (!_core.types.isStringLiteral(node, {
              value: ""
            ***REMOVED***)) {
              nodes.push(node);
            ***REMOVED***
          ***REMOVED***
        ***REMOVED***
        if (!_core.types.isStringLiteral(nodes[0]) && !(ignoreToPrimitiveHint && _core.types.isStringLiteral(nodes[1]))) {
          nodes.unshift(_core.types.stringLiteral(""));
        ***REMOVED***
        let root = nodes[0];
        if (ignoreToPrimitiveHint) {
          for (let i = 1; i < nodes.length; i++) {
            root = _core.types.binaryExpression("+", root, nodes[i]);
          ***REMOVED***
        ***REMOVED*** else if (nodes.length > 1) {
          root = buildConcatCallExpressions(nodes);
        ***REMOVED***
        path.replaceWith(root);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;
***REMOVED***);
