// @babel/helper-member-expression-to-functions/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true ***REMOVED***);

var _t = require('@babel/types');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; ***REMOVED***
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***
  n.default = e;
  return Object.freeze(n);
***REMOVED***

var _t__namespace = /*#__PURE__*/_interopNamespace(_t);

function willPathCastToBoolean(path) {
  const maybeWrapped = path;
  const {
    node,
    parentPath
  ***REMOVED*** = maybeWrapped;
  if (parentPath.isLogicalExpression()) {
    const {
      operator,
      right
    ***REMOVED*** = parentPath.node;
    if (operator === "&&" || operator === "||" || operator === "??" && node === right) {
      return willPathCastToBoolean(parentPath);
    ***REMOVED***
  ***REMOVED***
  if (parentPath.isSequenceExpression()) {
    const {
      expressions
    ***REMOVED*** = parentPath.node;
    if (expressions[expressions.length - 1] === node) {
      return willPathCastToBoolean(parentPath);
    ***REMOVED*** else {
      return true;
    ***REMOVED***
  ***REMOVED***
  return parentPath.isConditional({
    test: node
  ***REMOVED***) || parentPath.isUnaryExpression({
    operator: "!"
  ***REMOVED***) || parentPath.isLoop({
    test: node
  ***REMOVED***);
***REMOVED***

const {
  LOGICAL_OPERATORS,
  arrowFunctionExpression,
  assignmentExpression,
  binaryExpression,
  booleanLiteral,
  callExpression,
  cloneNode,
  conditionalExpression,
  identifier,
  isMemberExpression,
  isOptionalCallExpression,
  isOptionalMemberExpression,
  isUpdateExpression,
  logicalExpression,
  memberExpression,
  nullLiteral,
  optionalCallExpression,
  optionalMemberExpression,
  sequenceExpression,
  updateExpression
***REMOVED*** = _t__namespace;
class AssignmentMemoiser {
  constructor() {
    this._map = void 0;
    this._map = new WeakMap();
  ***REMOVED***
  has(key) {
    return this._map.has(key);
  ***REMOVED***
  get(key) {
    if (!this.has(key)) return;
    const record = this._map.get(key);
    const {
      value
    ***REMOVED*** = record;
    record.count--;
    if (record.count === 0) {
      return assignmentExpression("=", value, key);
    ***REMOVED***
    return value;
  ***REMOVED***
  set(key, value, count) {
    return this._map.set(key, {
      count,
      value
    ***REMOVED***);
  ***REMOVED***
***REMOVED***
function toNonOptional(path, base) {
  const {
    node
  ***REMOVED*** = path;
  if (isOptionalMemberExpression(node)) {
    return memberExpression(base, node.property, node.computed);
  ***REMOVED***
  if (path.isOptionalCallExpression()) {
    const callee = path.get("callee");
    if (path.node.optional && callee.isOptionalMemberExpression()) {
      const object = callee.node.object;
      const context = path.scope.maybeGenerateMemoised(object);
      callee.get("object").replaceWith(assignmentExpression("=", context, object));
      return callExpression(memberExpression(base, identifier("call")), [context, ...path.node.arguments]);
    ***REMOVED***
    return callExpression(base, path.node.arguments);
  ***REMOVED***
  return path.node;
***REMOVED***
function isInDetachedTree(path) {
  while (path) {
    if (path.isProgram()) break;
    const {
      parentPath,
      container,
      listKey
    ***REMOVED*** = path;
    const parentNode = parentPath.node;
    if (listKey) {
      if (container !== parentNode[listKey]) {
        return true;
      ***REMOVED***
    ***REMOVED*** else {
      if (container !== parentNode) return true;
    ***REMOVED***
    path = parentPath;
  ***REMOVED***
  return false;
***REMOVED***
const handle = {
  memoise() {***REMOVED***,
  handle(member, noDocumentAll) {
    const {
      node,
      parent,
      parentPath,
      scope
    ***REMOVED*** = member;
    if (member.isOptionalMemberExpression()) {
      if (isInDetachedTree(member)) return;
      const endPath = member.find(({
        node,
        parent
      ***REMOVED***) => {
        if (isOptionalMemberExpression(parent)) {
          return parent.optional || parent.object !== node;
        ***REMOVED***
        if (isOptionalCallExpression(parent)) {
          return (node !== member.node && parent.optional || parent.callee !== node
          );
        ***REMOVED***
        return true;
      ***REMOVED***);
      if (scope.path.isPattern()) {
        endPath.replaceWith(callExpression(arrowFunctionExpression([], endPath.node), []));
        return;
      ***REMOVED***
      const willEndPathCastToBoolean = willPathCastToBoolean(endPath);
      const rootParentPath = endPath.parentPath;
      if (rootParentPath.isUpdateExpression({
        argument: node
      ***REMOVED***)) {
        throw member.buildCodeFrameError(`can't handle update expression`);
      ***REMOVED***
      const isAssignment = rootParentPath.isAssignmentExpression({
        left: endPath.node
      ***REMOVED***);
      const isDeleteOperation = rootParentPath.isUnaryExpression({
        operator: "delete"
      ***REMOVED***);
      if (isDeleteOperation && endPath.isOptionalMemberExpression() && endPath.get("property").isPrivateName()) {
        throw member.buildCodeFrameError(`can't delete a private class element`);
      ***REMOVED***
      let startingOptional = member;
      for (;;) {
        if (startingOptional.isOptionalMemberExpression()) {
          if (startingOptional.node.optional) break;
          startingOptional = startingOptional.get("object");
          continue;
        ***REMOVED*** else if (startingOptional.isOptionalCallExpression()) {
          if (startingOptional.node.optional) break;
          startingOptional = startingOptional.get("callee");
          continue;
        ***REMOVED***
        throw new Error(`Internal error: unexpected ${startingOptional.node.type***REMOVED***`);
      ***REMOVED***
      const startingNode = startingOptional.isOptionalMemberExpression() ? startingOptional.node.object : startingOptional.node.callee;
      const baseNeedsMemoised = scope.maybeGenerateMemoised(startingNode);
      const baseRef = baseNeedsMemoised != null ? baseNeedsMemoised : startingNode;
      const parentIsOptionalCall = parentPath.isOptionalCallExpression({
        callee: node
      ***REMOVED***);
      const isOptionalCall = parent => parentIsOptionalCall;
      const parentIsCall = parentPath.isCallExpression({
        callee: node
      ***REMOVED***);
      startingOptional.replaceWith(toNonOptional(startingOptional, baseRef));
      if (isOptionalCall()) {
        if (parent.optional) {
          parentPath.replaceWith(this.optionalCall(member, parent.arguments));
        ***REMOVED*** else {
          parentPath.replaceWith(this.call(member, parent.arguments));
        ***REMOVED***
      ***REMOVED*** else if (parentIsCall) {
        member.replaceWith(this.boundGet(member));
      ***REMOVED*** else if (this.delete && parentPath.isUnaryExpression({
        operator: "delete"
      ***REMOVED***)) {
        parentPath.replaceWith(this.delete(member));
      ***REMOVED*** else if (parentPath.isAssignmentExpression()) {
        handleAssignment(this, member, parentPath);
      ***REMOVED*** else {
        member.replaceWith(this.get(member));
      ***REMOVED***
      let regular = member.node;
      for (let current = member; current !== endPath;) {
        const parentPath = current.parentPath;
        if (parentPath === endPath && isOptionalCall() && parent.optional) {
          regular = parentPath.node;
          break;
        ***REMOVED***
        regular = toNonOptional(parentPath, regular);
        current = parentPath;
      ***REMOVED***
      let context;
      const endParentPath = endPath.parentPath;
      if (isMemberExpression(regular) && endParentPath.isOptionalCallExpression({
        callee: endPath.node,
        optional: true
      ***REMOVED***)) {
        const {
          object
        ***REMOVED*** = regular;
        context = member.scope.maybeGenerateMemoised(object);
        if (context) {
          regular.object = assignmentExpression("=", context, object);
        ***REMOVED***
      ***REMOVED***
      let replacementPath = endPath;
      if (isDeleteOperation || isAssignment) {
        replacementPath = endParentPath;
        regular = endParentPath.node;
      ***REMOVED***
      const baseMemoised = baseNeedsMemoised ? assignmentExpression("=", cloneNode(baseRef), cloneNode(startingNode)) : cloneNode(baseRef);
      if (willEndPathCastToBoolean) {
        let nonNullishCheck;
        if (noDocumentAll) {
          nonNullishCheck = binaryExpression("!=", baseMemoised, nullLiteral());
        ***REMOVED*** else {
          nonNullishCheck = logicalExpression("&&", binaryExpression("!==", baseMemoised, nullLiteral()), binaryExpression("!==", cloneNode(baseRef), scope.buildUndefinedNode()));
        ***REMOVED***
        replacementPath.replaceWith(logicalExpression("&&", nonNullishCheck, regular));
      ***REMOVED*** else {
        let nullishCheck;
        if (noDocumentAll) {
          nullishCheck = binaryExpression("==", baseMemoised, nullLiteral());
        ***REMOVED*** else {
          nullishCheck = logicalExpression("||", binaryExpression("===", baseMemoised, nullLiteral()), binaryExpression("===", cloneNode(baseRef), scope.buildUndefinedNode()));
        ***REMOVED***
        replacementPath.replaceWith(conditionalExpression(nullishCheck, isDeleteOperation ? booleanLiteral(true) : scope.buildUndefinedNode(), regular));
      ***REMOVED***
      if (context) {
        const endParent = endParentPath.node;
        endParentPath.replaceWith(optionalCallExpression(optionalMemberExpression(endParent.callee, identifier("call"), false, true), [cloneNode(context), ...endParent.arguments], false));
      ***REMOVED***
      return;
    ***REMOVED***
    if (isUpdateExpression(parent, {
      argument: node
    ***REMOVED***)) {
      if (this.simpleSet) {
        member.replaceWith(this.simpleSet(member));
        return;
      ***REMOVED***
      const {
        operator,
        prefix
      ***REMOVED*** = parent;
      this.memoise(member, 2);
      const ref = scope.generateUidIdentifierBasedOnNode(node);
      scope.push({
        id: ref
      ***REMOVED***);
      const seq = [assignmentExpression("=", cloneNode(ref), this.get(member))];
      if (prefix) {
        seq.push(updateExpression(operator, cloneNode(ref), prefix));
        const value = sequenceExpression(seq);
        parentPath.replaceWith(this.set(member, value));
        return;
      ***REMOVED*** else {
        const ref2 = scope.generateUidIdentifierBasedOnNode(node);
        scope.push({
          id: ref2
        ***REMOVED***);
        seq.push(assignmentExpression("=", cloneNode(ref2), updateExpression(operator, cloneNode(ref), prefix)), cloneNode(ref));
        const value = sequenceExpression(seq);
        parentPath.replaceWith(sequenceExpression([this.set(member, value), cloneNode(ref2)]));
        return;
      ***REMOVED***
    ***REMOVED***
    if (parentPath.isAssignmentExpression({
      left: node
    ***REMOVED***)) {
      handleAssignment(this, member, parentPath);
      return;
    ***REMOVED***
    if (parentPath.isCallExpression({
      callee: node
    ***REMOVED***)) {
      parentPath.replaceWith(this.call(member, parentPath.node.arguments));
      return;
    ***REMOVED***
    if (parentPath.isOptionalCallExpression({
      callee: node
    ***REMOVED***)) {
      if (scope.path.isPattern()) {
        parentPath.replaceWith(callExpression(arrowFunctionExpression([], parentPath.node), []));
        return;
      ***REMOVED***
      parentPath.replaceWith(this.optionalCall(member, parentPath.node.arguments));
      return;
    ***REMOVED***
    if (this.delete && parentPath.isUnaryExpression({
      operator: "delete"
    ***REMOVED***)) {
      parentPath.replaceWith(this.delete(member));
      return;
    ***REMOVED***
    if (parentPath.isForXStatement({
      left: node
    ***REMOVED***) || parentPath.isObjectProperty({
      value: node
    ***REMOVED***) && parentPath.parentPath.isObjectPattern() || parentPath.isAssignmentPattern({
      left: node
    ***REMOVED***) && parentPath.parentPath.isObjectProperty({
      value: parent
    ***REMOVED***) && parentPath.parentPath.parentPath.isObjectPattern() || parentPath.isArrayPattern() || parentPath.isAssignmentPattern({
      left: node
    ***REMOVED***) && parentPath.parentPath.isArrayPattern() || parentPath.isRestElement()) {
      member.replaceWith(this.destructureSet(member));
      return;
    ***REMOVED***
    if (parentPath.isTaggedTemplateExpression()) {
      member.replaceWith(this.boundGet(member));
    ***REMOVED*** else {
      member.replaceWith(this.get(member));
    ***REMOVED***
  ***REMOVED***
***REMOVED***;
function handleAssignment(state, member, parentPath) {
  if (state.simpleSet) {
    member.replaceWith(state.simpleSet(member));
    return;
  ***REMOVED***
  const {
    operator,
    right: value
  ***REMOVED*** = parentPath.node;
  if (operator === "=") {
    parentPath.replaceWith(state.set(member, value));
  ***REMOVED*** else {
    const operatorTrunc = operator.slice(0, -1);
    if (LOGICAL_OPERATORS.includes(operatorTrunc)) {
      state.memoise(member, 1);
      parentPath.replaceWith(logicalExpression(operatorTrunc, state.get(member), state.set(member, value)));
    ***REMOVED*** else {
      state.memoise(member, 2);
      parentPath.replaceWith(state.set(member, binaryExpression(operatorTrunc, state.get(member), value)));
    ***REMOVED***
  ***REMOVED***
***REMOVED***
function memberExpressionToFunctions(path, visitor, state) {
  path.traverse(visitor, Object.assign({***REMOVED***, handle, state, {
    memoiser: new AssignmentMemoiser()
  ***REMOVED***));
***REMOVED***

exports.default = memberExpressionToFunctions;
