// jsdom/lib/jsdom/browser/resources/resource-queue.js
"use strict";

/**
 * Queue for all the resources to be download except async scripts.
 * Async scripts have their own queue AsyncResourceQueue.
 */
module.exports = class ResourceQueue {
  constructor({ paused, asyncQueue ***REMOVED*** = {***REMOVED***) {
    this.paused = Boolean(paused);
    this._asyncQueue = asyncQueue;
  ***REMOVED***

  getLastScript() {
    let head = this.tail;

    while (head) {
      if (head.isScript) {
        return head;
      ***REMOVED***
      head = head.prev;
    ***REMOVED***

    return null;
  ***REMOVED***

  _moreScripts() {
    let found = false;

    let head = this.tail;
    while (head && !found) {
      found = head.isScript;
      head = head.prev;
    ***REMOVED***

    return found;
  ***REMOVED***

  _notify() {
    if (this._listener) {
      this._listener();
    ***REMOVED***
  ***REMOVED***

  setListener(listener) {
    this._listener = listener;
  ***REMOVED***

  push(request, onLoad, onError, keepLast, element) {
    const isScript = element ? element.localName === "script" : false;

    if (!request) {
      if (isScript && !this._moreScripts()) {
        return onLoad();
      ***REMOVED***

      request = Promise.resolve();
    ***REMOVED***
    const q = this;
    const item = {
      isScript,
      err: null,
      element,
      fired: false,
      data: null,
      keepLast,
      prev: q.tail,
      check() {
        if (!q.paused && !this.prev && this.fired) {
          let promise;

          if (this.err && onError) {
            promise = onError(this.err);
          ***REMOVED***

          if (!this.err && onLoad) {
            promise = onLoad(this.data);
          ***REMOVED***

          Promise.resolve(promise)
            .then(() => {
              if (this.next) {
                this.next.prev = null;
                this.next.check();
              ***REMOVED*** else { // q.tail===this
                q.tail = null;
                q._notify();
              ***REMOVED***

              this.finished = true;

              if (q._asyncQueue) {
                q._asyncQueue.notifyItem(this);
              ***REMOVED***
            ***REMOVED***);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***;
    if (q.tail) {
      if (q.tail.keepLast) {
        // if the tail is the load event in document and we receive a new element to load
        // we should add this new request before the load event.
        if (q.tail.prev) {
          q.tail.prev.next = item;
        ***REMOVED***
        item.prev = q.tail.prev;
        q.tail.prev = item;
        item.next = q.tail;
      ***REMOVED*** else {
        q.tail.next = item;
        q.tail = item;
      ***REMOVED***
    ***REMOVED*** else {
      q.tail = item;
    ***REMOVED***
    return request
      .then(data => {
        item.fired = 1;
        item.data = data;
        item.check();
      ***REMOVED***)
      .catch(err => {
        item.fired = true;
        item.err = err;
        item.check();
      ***REMOVED***);
  ***REMOVED***

  resume() {
    if (!this.paused) {
      return;
    ***REMOVED***
    this.paused = false;

    let head = this.tail;
    while (head && head.prev) {
      head = head.prev;
    ***REMOVED***
    if (head) {
      head.check();
    ***REMOVED***
  ***REMOVED***
***REMOVED***;
