// @humanwhocodes/object-schema/src/validation-strategy.js
/**
 * @filedescription Validation Strategy
 */

"use strict";

//-----------------------------------------------------------------------------
// Class
//-----------------------------------------------------------------------------

/**
 * Container class for several different validation strategies.
 */
class ValidationStrategy {

    /**
     * Validates that a value is an array.
     * @param {****REMOVED*** value The value to validate.
     * @returns {void***REMOVED***
     * @throws {TypeError***REMOVED*** If the value is invalid. 
     */
    static array(value) {
        if (!Array.isArray(value)) {
            throw new TypeError("Expected an array.");
        ***REMOVED***
    ***REMOVED***

    /**
     * Validates that a value is a boolean.
     * @param {****REMOVED*** value The value to validate.
     * @returns {void***REMOVED***
     * @throws {TypeError***REMOVED*** If the value is invalid. 
     */
    static boolean(value) {
        if (typeof value !== "boolean") {
            throw new TypeError("Expected a Boolean.");
        ***REMOVED***
    ***REMOVED***

    /**
     * Validates that a value is a number.
     * @param {****REMOVED*** value The value to validate.
     * @returns {void***REMOVED***
     * @throws {TypeError***REMOVED*** If the value is invalid. 
     */
    static number(value) {
        if (typeof value !== "number") {
            throw new TypeError("Expected a number.");
        ***REMOVED***
    ***REMOVED***

    /**
     * Validates that a value is a object.
     * @param {****REMOVED*** value The value to validate.
     * @returns {void***REMOVED***
     * @throws {TypeError***REMOVED*** If the value is invalid. 
     */
    static object(value) {
        if (!value || typeof value !== "object") {
            throw new TypeError("Expected an object.");
        ***REMOVED***
    ***REMOVED***

    /**
     * Validates that a value is a object or null.
     * @param {****REMOVED*** value The value to validate.
     * @returns {void***REMOVED***
     * @throws {TypeError***REMOVED*** If the value is invalid. 
     */
    static "object?"(value) {
        if (typeof value !== "object") {
            throw new TypeError("Expected an object or null.");
        ***REMOVED***
    ***REMOVED***

    /**
     * Validates that a value is a string.
     * @param {****REMOVED*** value The value to validate.
     * @returns {void***REMOVED***
     * @throws {TypeError***REMOVED*** If the value is invalid. 
     */
    static string(value) {
        if (typeof value !== "string") {
            throw new TypeError("Expected a string.");
        ***REMOVED***
    ***REMOVED***

    /**
     * Validates that a value is a non-empty string.
     * @param {****REMOVED*** value The value to validate.
     * @returns {void***REMOVED***
     * @throws {TypeError***REMOVED*** If the value is invalid. 
     */
    static "string!"(value) {
        if (typeof value !== "string" || value.length === 0) {
            throw new TypeError("Expected a non-empty string.");
        ***REMOVED***
    ***REMOVED***

***REMOVED***

exports.ValidationStrategy = ValidationStrategy;
