// @babel/eslint-parser/lib/analyze-scope.cjs
"use strict";

function _classPrivateFieldInitSpec(e, t, a) { _checkPrivateRedeclaration(e, t), t.set(e, a); ***REMOVED***
function _checkPrivateRedeclaration(e, t) { if (t.has(e)) throw new TypeError("Cannot initialize the same private elements twice on an object"); ***REMOVED***
function _classPrivateFieldGet(s, a) { return s.get(_assertClassBrand(s, a)); ***REMOVED***
function _classPrivateFieldSet(s, a, r) { return s.set(_assertClassBrand(s, a), r), r; ***REMOVED***
function _assertClassBrand(e, t, n) { if ("function" == typeof e ? e === t : e.has(t)) return arguments.length < 3 ? t : n; throw new TypeError("Private element is not present on this object"); ***REMOVED***
const {
  Definition,
  PatternVisitor: OriginalPatternVisitor,
  Referencer: OriginalReferencer,
  Scope,
  ScopeManager
***REMOVED*** = require("@nicolo-ribaudo/eslint-scope-5-internals");
const {
  getKeys: fallback
***REMOVED*** = require("eslint-visitor-keys");
let visitorKeysMap;
function getVisitorValues(nodeType, client) {
  if (visitorKeysMap) return visitorKeysMap[nodeType];
  const {
    FLOW_FLIPPED_ALIAS_KEYS,
    VISITOR_KEYS
  ***REMOVED*** = client.getTypesInfo();
  const flowFlippedAliasKeys = new Set(FLOW_FLIPPED_ALIAS_KEYS.concat(["ArrayPattern", "ClassDeclaration", "ClassExpression", "FunctionDeclaration", "FunctionExpression", "Identifier", "ObjectPattern", "RestElement"]));
  visitorKeysMap = (Object.entries || (o => Object.keys(o).map(k => [k, o[k]])))(VISITOR_KEYS).reduce((acc, [key, value]) => {
    if (!flowFlippedAliasKeys.has(value)) {
      acc[key] = value;
    ***REMOVED***
    return acc;
  ***REMOVED***, {***REMOVED***);
  return visitorKeysMap[nodeType];
***REMOVED***
const propertyTypes = {
  callProperties: {
    type: "loop",
    values: ["value"]
  ***REMOVED***,
  indexers: {
    type: "loop",
    values: ["key", "value"]
  ***REMOVED***,
  properties: {
    type: "loop",
    values: ["argument", "value"]
  ***REMOVED***,
  types: {
    type: "loop"
  ***REMOVED***,
  params: {
    type: "loop"
  ***REMOVED***,
  argument: {
    type: "single"
  ***REMOVED***,
  elementType: {
    type: "single"
  ***REMOVED***,
  qualification: {
    type: "single"
  ***REMOVED***,
  rest: {
    type: "single"
  ***REMOVED***,
  returnType: {
    type: "single"
  ***REMOVED***,
  typeAnnotation: {
    type: "typeAnnotation"
  ***REMOVED***,
  typeParameters: {
    type: "typeParameters"
  ***REMOVED***,
  id: {
    type: "id"
  ***REMOVED***
***REMOVED***;
class PatternVisitor extends OriginalPatternVisitor {
  ArrayPattern(node) {
    node.elements.forEach(this.visit, this);
  ***REMOVED***
  ObjectPattern(node) {
    node.properties.forEach(this.visit, this);
  ***REMOVED***
***REMOVED***
var _client = new WeakMap();
class Referencer extends OriginalReferencer {
  constructor(options, scopeManager, client) {
    super(options, scopeManager);
    _classPrivateFieldInitSpec(this, _client, void 0);
    _classPrivateFieldSet(_client, this, client);
  ***REMOVED***
  visitPattern(node, options, callback) {
    if (!node) {
      return;
    ***REMOVED***
    this._checkIdentifierOrVisit(node.typeAnnotation);
    if (node.type === "AssignmentPattern") {
      this._checkIdentifierOrVisit(node.left.typeAnnotation);
    ***REMOVED***
    if (typeof options === "function") {
      callback = options;
      options = {
        processRightHandNodes: false
      ***REMOVED***;
    ***REMOVED***
    const visitor = new PatternVisitor(this.options, node, callback);
    visitor.visit(node);
    if (options.processRightHandNodes) {
      visitor.rightHandNodes.forEach(this.visit, this);
    ***REMOVED***
  ***REMOVED***
  visitClass(node) {
    var _ref;
    this._visitArray(node.decorators);
    const typeParamScope = this._nestTypeParamScope(node);
    this._visitTypeAnnotation(node.implements);
    this._visitTypeAnnotation((_ref = node.superTypeParameters) == null ? void 0 : _ref.params);
    super.visitClass(node);
    if (typeParamScope) {
      this.close(node);
    ***REMOVED***
  ***REMOVED***
  visitFunction(node) {
    const typeParamScope = this._nestTypeParamScope(node);
    this._checkIdentifierOrVisit(node.returnType);
    super.visitFunction(node);
    if (typeParamScope) {
      this.close(node);
    ***REMOVED***
  ***REMOVED***
  visitProperty(node) {
    var _node$value;
    if (((_node$value = node.value) == null ? void 0 : _node$value.type) === "TypeCastExpression") {
      this._visitTypeAnnotation(node.value);
    ***REMOVED***
    this._visitArray(node.decorators);
    super.visitProperty(node);
  ***REMOVED***
  InterfaceDeclaration(node) {
    this._createScopeVariable(node, node.id);
    const typeParamScope = this._nestTypeParamScope(node);
    this._visitArray(node.extends);
    this.visit(node.body);
    if (typeParamScope) {
      this.close(node);
    ***REMOVED***
  ***REMOVED***
  TypeAlias(node) {
    this._createScopeVariable(node, node.id);
    const typeParamScope = this._nestTypeParamScope(node);
    this.visit(node.right);
    if (typeParamScope) {
      this.close(node);
    ***REMOVED***
  ***REMOVED***
  ClassProperty(node) {
    this._visitClassProperty(node);
  ***REMOVED***
  ClassPrivateProperty(node) {
    this._visitClassProperty(node);
  ***REMOVED***
  AccessorProperty(node) {
    this._visitClassProperty(node);
  ***REMOVED***
  ClassAccessorProperty(node) {
    this._visitClassProperty(node);
  ***REMOVED***
  PropertyDefinition(node) {
    this._visitClassProperty(node);
  ***REMOVED***
  ClassPrivateMethod(node) {
    super.MethodDefinition(node);
  ***REMOVED***
  DeclareModule(node) {
    this._visitDeclareX(node);
  ***REMOVED***
  DeclareFunction(node) {
    this._visitDeclareX(node);
  ***REMOVED***
  DeclareVariable(node) {
    this._visitDeclareX(node);
  ***REMOVED***
  DeclareClass(node) {
    this._visitDeclareX(node);
  ***REMOVED***
  OptionalMemberExpression(node) {
    super.MemberExpression(node);
  ***REMOVED***
  _visitClassProperty(node) {
    const {
      computed,
      key,
      typeAnnotation,
      decorators,
      value
    ***REMOVED*** = node;
    this._visitArray(decorators);
    if (computed) this.visit(key);
    this._visitTypeAnnotation(typeAnnotation);
    if (value) {
      if (this.scopeManager.__nestClassFieldInitializerScope) {
        this.scopeManager.__nestClassFieldInitializerScope(value);
      ***REMOVED*** else {
        this.scopeManager.__nestScope(new Scope(this.scopeManager, "function", this.scopeManager.__currentScope, value, true));
      ***REMOVED***
      this.visit(value);
      this.close(value);
    ***REMOVED***
  ***REMOVED***
  _visitDeclareX(node) {
    if (node.id) {
      this._createScopeVariable(node, node.id);
    ***REMOVED***
    const typeParamScope = this._nestTypeParamScope(node);
    if (typeParamScope) {
      this.close(node);
    ***REMOVED***
  ***REMOVED***
  _createScopeVariable(node, name) {
    this.currentScope().variableScope.__define(name, new Definition("Variable", name, node, null, null, null));
  ***REMOVED***
  _nestTypeParamScope(node) {
    if (!node.typeParameters) {
      return null;
    ***REMOVED***
    const parentScope = this.scopeManager.__currentScope;
    const scope = new Scope(this.scopeManager, "type-parameters", parentScope, node, false);
    this.scopeManager.__nestScope(scope);
    for (let j = 0; j < node.typeParameters.params.length; j++) {
      const name = node.typeParameters.params[j];
      scope.__define(name, new Definition("TypeParameter", name, name));
      if (name.typeAnnotation) {
        this._checkIdentifierOrVisit(name);
      ***REMOVED***
    ***REMOVED***
    scope.__define = parentScope.__define.bind(parentScope);
    return scope;
  ***REMOVED***
  _visitTypeAnnotation(node) {
    if (!node) {
      return;
    ***REMOVED***
    if (Array.isArray(node)) {
      node.forEach(this._visitTypeAnnotation, this);
      return;
    ***REMOVED***
    const visitorValues = getVisitorValues(node.type, _classPrivateFieldGet(_client, this));
    if (!visitorValues) {
      return;
    ***REMOVED***
    for (let i = 0; i < visitorValues.length; i++) {
      const visitorValue = visitorValues[i];
      const propertyType = propertyTypes[visitorValue];
      const nodeProperty = node[visitorValue];
      if (propertyType == null || nodeProperty == null) {
        continue;
      ***REMOVED***
      if (propertyType.type === "loop") {
        for (let j = 0; j < nodeProperty.length; j++) {
          if (Array.isArray(propertyType.values)) {
            for (let k = 0; k < propertyType.values.length; k++) {
              const loopPropertyNode = nodeProperty[j][propertyType.values[k]];
              if (loopPropertyNode) {
                this._checkIdentifierOrVisit(loopPropertyNode);
              ***REMOVED***
            ***REMOVED***
          ***REMOVED*** else {
            this._checkIdentifierOrVisit(nodeProperty[j]);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED*** else if (propertyType.type === "single") {
        this._checkIdentifierOrVisit(nodeProperty);
      ***REMOVED*** else if (propertyType.type === "typeAnnotation") {
        this._visitTypeAnnotation(node.typeAnnotation);
      ***REMOVED*** else if (propertyType.type === "typeParameters") {
        for (let l = 0; l < node.typeParameters.params.length; l++) {
          this._checkIdentifierOrVisit(node.typeParameters.params[l]);
        ***REMOVED***
      ***REMOVED*** else if (propertyType.type === "id") {
        if (node.id.type === "Identifier") {
          this._checkIdentifierOrVisit(node.id);
        ***REMOVED*** else {
          this._visitTypeAnnotation(node.id);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  _checkIdentifierOrVisit(node) {
    if (node != null && node.typeAnnotation) {
      this._visitTypeAnnotation(node.typeAnnotation);
    ***REMOVED*** else if ((node == null ? void 0 : node.type) === "Identifier") {
      this.visit(node);
    ***REMOVED*** else {
      this._visitTypeAnnotation(node);
    ***REMOVED***
  ***REMOVED***
  _visitArray(nodeList) {
    if (nodeList) {
      for (const node of nodeList) {
        this.visit(node);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***
module.exports = function analyzeScope(ast, parserOptions, client) {
  var _parserOptions$ecmaFe;
  const options = {
    ignoreEval: true,
    optimistic: false,
    directive: false,
    nodejsScope: ast.sourceType === "script" && ((_parserOptions$ecmaFe = parserOptions.ecmaFeatures) == null ? void 0 : _parserOptions$ecmaFe.globalReturn) === true,
    impliedStrict: false,
    sourceType: ast.sourceType,
    ecmaVersion: parserOptions.ecmaVersion,
    fallback,
    childVisitorKeys: client.getVisitorKeys()
  ***REMOVED***;
  const scopeManager = new ScopeManager(options);
  const referencer = new Referencer(options, scopeManager, client);
  referencer.visit(ast);
  return scopeManager;
***REMOVED***;
