// ajv-keywords/keywords/dynamicDefaults.js
'use strict';

var sequences = {***REMOVED***;

var DEFAULTS = {
  timestamp: function() { return Date.now(); ***REMOVED***,
  datetime: function() { return (new Date).toISOString(); ***REMOVED***,
  date: function() { return (new Date).toISOString().slice(0, 10); ***REMOVED***,
  time: function() { return (new Date).toISOString().slice(11); ***REMOVED***,
  random: function() { return Math.random(); ***REMOVED***,
  randomint: function (args) {
    var limit = args && args.max || 2;
    return function() { return Math.floor(Math.random() * limit); ***REMOVED***;
  ***REMOVED***,
  seq: function (args) {
    var name = args && args.name || '';
    sequences[name] = sequences[name] || 0;
    return function() { return sequences[name]++; ***REMOVED***;
  ***REMOVED***
***REMOVED***;

module.exports = function defFunc(ajv) {
  defFunc.definition = {
    compile: function (schema, parentSchema, it) {
      var funcs = {***REMOVED***;

      for (var key in schema) {
        var d = schema[key];
        var func = getDefault(typeof d == 'string' ? d : d.func);
        funcs[key] = func.length ? func(d.args) : func;
      ***REMOVED***

      return it.opts.useDefaults && !it.compositeRule
              ? assignDefaults
              : noop;

      function assignDefaults(data) {
        for (var prop in schema){
          if (data[prop] === undefined
            || (it.opts.useDefaults == 'empty'
            && (data[prop] === null || data[prop] === '')))
            data[prop] = funcs[prop]();
        ***REMOVED***
        return true;
      ***REMOVED***

      function noop() { return true; ***REMOVED***
    ***REMOVED***,
    DEFAULTS: DEFAULTS,
    metaSchema: {
      type: 'object',
      additionalProperties: {
        type: ['string', 'object'],
        additionalProperties: false,
        required: ['func', 'args'],
        properties: {
          func: { type: 'string' ***REMOVED***,
          args: { type: 'object' ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;

  ajv.addKeyword('dynamicDefaults', defFunc.definition);
  return ajv;

  function getDefault(d) {
    var def = DEFAULTS[d];
    if (def) return def;
    throw new Error('invalid "dynamicDefaults" keyword property value: ' + d);
  ***REMOVED***
***REMOVED***;
