// ajv-keywords/keywords/regexp.js
'use strict';

module.exports = function defFunc(ajv) {
  defFunc.definition = {
    type: 'string',
    inline: function (it, keyword, schema) {
      return getRegExp() + '.test(data' + (it.dataLevel || '') + ')';

      function getRegExp() {
        try {
          if (typeof schema == 'object')
            return new RegExp(schema.pattern, schema.flags);

          var rx = schema.match(/^\/(.*)\/([gimuy]*)$/);
          if (rx) return new RegExp(rx[1], rx[2]);
          throw new Error('cannot parse string into RegExp');
        ***REMOVED*** catch(e) {
          console.error('regular expression', schema, 'is invalid');
          throw e;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***,
    metaSchema: {
      type: ['string', 'object'],
      properties: {
        pattern: { type: 'string' ***REMOVED***,
        flags: { type: 'string' ***REMOVED***
      ***REMOVED***,
      required: ['pattern'],
      additionalProperties: false
    ***REMOVED***
  ***REMOVED***;

  ajv.addKeyword('regexp', defFunc.definition);
  return ajv;
***REMOVED***;
