// webpack/lib/javascript/EnableChunkLoadingPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

/** @typedef {import("../../declarations/WebpackOptions").ChunkLoadingType***REMOVED*** ChunkLoadingType */
/** @typedef {import("../Compiler")***REMOVED*** Compiler */

/** @type {WeakMap<Compiler, Set<ChunkLoadingType>>***REMOVED*** */
const enabledTypes = new WeakMap();

/**
 * @param {Compiler***REMOVED*** compiler compiler
 * @returns {Set<ChunkLoadingType>***REMOVED*** enabled types
 */
const getEnabledTypes = compiler => {
	let set = enabledTypes.get(compiler);
	if (set === undefined) {
		set = new Set();
		enabledTypes.set(compiler, set);
	***REMOVED***
	return set;
***REMOVED***;

class EnableChunkLoadingPlugin {
	/**
	 * @param {ChunkLoadingType***REMOVED*** type library type that should be available
	 */
	constructor(type) {
		this.type = type;
	***REMOVED***

	/**
	 * @param {Compiler***REMOVED*** compiler the compiler instance
	 * @param {ChunkLoadingType***REMOVED*** type type of library
	 * @returns {void***REMOVED***
	 */
	static setEnabled(compiler, type) {
		getEnabledTypes(compiler).add(type);
	***REMOVED***

	/**
	 * @param {Compiler***REMOVED*** compiler the compiler instance
	 * @param {ChunkLoadingType***REMOVED*** type type of library
	 * @returns {void***REMOVED***
	 */
	static checkEnabled(compiler, type) {
		if (!getEnabledTypes(compiler).has(type)) {
			throw new Error(
				`Chunk loading type "${type***REMOVED***" is not enabled. ` +
					"EnableChunkLoadingPlugin need to be used to enable this type of chunk loading. " +
					'This usually happens through the "output.enabledChunkLoadingTypes" option. ' +
					'If you are using a function as entry which sets "chunkLoading", you need to add all potential chunk loading types to "output.enabledChunkLoadingTypes". ' +
					`These types are enabled: ${Array.from(
						getEnabledTypes(compiler)
					).join(", ")***REMOVED***`
			);
		***REMOVED***
	***REMOVED***

	/**
	 * Apply the plugin
	 * @param {Compiler***REMOVED*** compiler the compiler instance
	 * @returns {void***REMOVED***
	 */
	apply(compiler) {
		const { type ***REMOVED*** = this;

		// Only enable once
		const enabled = getEnabledTypes(compiler);
		if (enabled.has(type)) return;
		enabled.add(type);

		if (typeof type === "string") {
			switch (type) {
				case "jsonp": {
					const JsonpChunkLoadingPlugin = require("../web/JsonpChunkLoadingPlugin");
					new JsonpChunkLoadingPlugin().apply(compiler);
					break;
				***REMOVED***
				case "import-scripts": {
					const ImportScriptsChunkLoadingPlugin = require("../webworker/ImportScriptsChunkLoadingPlugin");
					new ImportScriptsChunkLoadingPlugin().apply(compiler);
					break;
				***REMOVED***
				case "require": {
					// @ts-expect-error https://github.com/microsoft/TypeScript/issues/41697
					const CommonJsChunkLoadingPlugin = require("../node/CommonJsChunkLoadingPlugin");
					new CommonJsChunkLoadingPlugin({
						asyncChunkLoading: false
					***REMOVED***).apply(compiler);
					break;
				***REMOVED***
				case "async-node": {
					// @ts-expect-error https://github.com/microsoft/TypeScript/issues/41697
					const CommonJsChunkLoadingPlugin = require("../node/CommonJsChunkLoadingPlugin");
					new CommonJsChunkLoadingPlugin({
						asyncChunkLoading: true
					***REMOVED***).apply(compiler);
					break;
				***REMOVED***
				case "import":
				case "universal": {
					const ModuleChunkLoadingPlugin = require("../esm/ModuleChunkLoadingPlugin");
					new ModuleChunkLoadingPlugin().apply(compiler);
					break;
				***REMOVED***
				default:
					throw new Error(`Unsupported chunk loading type ${type***REMOVED***.
Plugins which provide custom chunk loading types must call EnableChunkLoadingPlugin.setEnabled(compiler, type) to disable this error.`);
			***REMOVED***
		***REMOVED*** else {
			// TODO support plugin instances here
			// apply them to the compiler
		***REMOVED***
	***REMOVED***
***REMOVED***

module.exports = EnableChunkLoadingPlugin;
