// resolve-url-loader/lib/engine/rework.js
/*
 * MIT License http://opensource.org/licenses/MIT
 * Author: Ben Holloway @bholloway
 */
'use strict';

var path    = require('path'),
    convert = require('convert-source-map');

var fileProtocol = require('../file-protocol');

var rework = requireOptionalPeerDependency('rework'),
    visit  = requireOptionalPeerDependency('rework-visit');

/**
 * Process the given CSS content into reworked CSS content.
 *
 * @param {string***REMOVED*** sourceFile The absolute path of the file being processed
 * @param {string***REMOVED*** sourceContent CSS content without source-map
 * @param {{outputSourceMap: boolean, transformDeclaration:function, absSourceMap:object,
 *        sourceMapConsumer:object***REMOVED******REMOVED*** params Named parameters
 * @return {{content: string, map: object***REMOVED******REMOVED*** Reworked CSS and optional source-map
 */
function process(sourceFile, sourceContent, params) {

  // embed source-map in css
  //  prepend file protocol to all sources to avoid problems with source map
  var contentWithMap = sourceContent + (
    params.absSourceMap ?
      convert.fromObject(fileProtocol.prepend(params.absSourceMap)).toComment({multiline: true***REMOVED***) :
      ''
  );

  // need to prepend file protocol to source as well to avoid problems with source map
  var reworked = rework(contentWithMap, {source: fileProtocol.prepend(sourceFile)***REMOVED***)
    .use(reworkPlugin)
    .toString({
      sourcemap        : params.outputSourceMap,
      sourcemapAsObject: params.outputSourceMap
    ***REMOVED***);

  // complete with source-map
  if (params.outputSourceMap) {
    return {
      content: reworked.code,
      map    : fileProtocol.remove(reworked.map)
    ***REMOVED***;
  ***REMOVED***
  // complete without source-map
  else {
    return {
      content: reworked,
      map    : null
    ***REMOVED***;
  ***REMOVED***

  /**
   * Plugin for css rework that follows SASS transpilation.
   *
   * @param {object***REMOVED*** stylesheet AST for the CSS output from SASS
   */
  function reworkPlugin(stylesheet) {

    // visit each node (selector) in the stylesheet recursively using the official utility method
    //  each node may have multiple declarations
    visit(stylesheet, function visitor(declarations) {
      if (declarations) {
        declarations.forEach(eachDeclaration);
      ***REMOVED***
    ***REMOVED***);

    /**
     * Process a declaration from the syntax tree.
     * @param declaration
     */
    function eachDeclaration(declaration) {
      var isValid = declaration.value && (declaration.value.indexOf('url') >= 0);
      if (isValid) {
        declaration.value = params.transformDeclaration(declaration.value, getPathsAtChar);
      ***REMOVED***

      /**
       * Create a hash of base path strings.
       *
       * Position in the declaration is not supported since rework does not refine sourcemaps to this detail.
       *
       * @throws Error on invalid source map
       * @returns {{selector:string, property:string***REMOVED******REMOVED*** Hash of base path strings
       */
      function getPathsAtChar() {
        var posSelector = declaration.parent && declaration.parent.position.start,
            posProperty = declaration.position.start;

        var result = {
          property: positionToOriginalDirectory(posProperty),
          selector: positionToOriginalDirectory(posSelector)
        ***REMOVED***;

        var isValid = [result.property, result.selector].every(Boolean);
        if (isValid) {
          return result;
        ***REMOVED***
        else if (params.sourceMapConsumer) {
          throw new Error('source-map information is not available at url() declaration');
        ***REMOVED*** else {
          throw new Error('a valid source-map is not present (ensure preceding loaders output a source-map)');
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  /**
   * Given an apparent position find the directory of the original file.
   *
   * @param startPosApparent {{line: number, column: number***REMOVED******REMOVED***
   * @returns {false|string***REMOVED*** Directory of original file or false on invalid
   */
  function positionToOriginalDirectory(startPosApparent) {
    // reverse the original source-map to find the original source file before transpilation
    var startPosOriginal =
      !!params.sourceMapConsumer &&
      params.sourceMapConsumer.originalPositionFor(startPosApparent);

    // we require a valid directory for the specified file
    var directory =
      !!startPosOriginal &&
      !!startPosOriginal.source &&
      fileProtocol.remove(path.dirname(startPosOriginal.source));

    return directory;
  ***REMOVED***
***REMOVED***

module.exports = process;

/**
 * Require the given filename but fail with an error that `requireOptionalPeerDependencies` must be installed.
 *
 * @param moduleName The module to require
 * @returns {****REMOVED*** The module
 * @throws Error when module is not found
 */
function requireOptionalPeerDependency(moduleName) {
  try {
    return require(moduleName);
  ***REMOVED***
  catch (error) {
    if (error.message === 'Cannot find module \'' + moduleName + '\'') {
      throw new Error('to use the "rework" engine you must install the optionalPeerDependencies');
    ***REMOVED***
    else {
      throw error;
    ***REMOVED***
  ***REMOVED***
***REMOVED***
