// jake/lib/namespace.js
const ROOT_NAMESPACE_NAME = '__rootNamespace__';

class Namespace {
  constructor(name, parentNamespace) {
    this.name = name;
    this.parentNamespace = parentNamespace;
    this.childNamespaces = {***REMOVED***;
    this.tasks = {***REMOVED***;
    this.rules = {***REMOVED***;
    this.path = this.getPath();
  ***REMOVED***

  get fullName() {
    return this._getFullName();
  ***REMOVED***

  addTask(task) {
    this.tasks[task.name] = task;
    task.namespace = this;
  ***REMOVED***

  resolveTask(name) {
    if (!name) {
      return;
    ***REMOVED***

    let taskPath = name.split(':');
    let taskName = taskPath.pop();
    let task;
    let ns;

    // Namespaced, return either relative to current, or from root
    if (taskPath.length) {
      taskPath = taskPath.join(':');
      ns = this.resolveNamespace(taskPath) ||
        Namespace.ROOT_NAMESPACE.resolveNamespace(taskPath);
      task = (ns && ns.resolveTask(taskName));
    ***REMOVED***
    // Bare task, return either local, or top-level
    else {
      task = this.tasks[name] || Namespace.ROOT_NAMESPACE.tasks[name];
    ***REMOVED***

    return task || null;
  ***REMOVED***


  resolveNamespace(relativeName) {
    if (!relativeName) {
      return this;
    ***REMOVED***

    let parts = relativeName.split(':');
    let ns = this;

    for (let i = 0, ii = parts.length; (ns && i < ii); i++) {
      ns = ns.childNamespaces[parts[i]];
    ***REMOVED***

    return ns || null;
  ***REMOVED***

  matchRule(relativeName) {
    let parts = relativeName.split(':');
    parts.pop();
    let ns = this.resolveNamespace(parts.join(':'));
    let rules = ns ? ns.rules : [];
    let r;
    let match;

    for (let p in rules) {
      r = rules[p];
      if (r.match(relativeName)) {
        match = r;
      ***REMOVED***
    ***REMOVED***

    return (ns && match) ||
        (this.parentNamespace &&
        this.parentNamespace.matchRule(relativeName));
  ***REMOVED***

  getPath() {
    let parts = [];
    let next = this.parentNamespace;
    while (next) {
      parts.push(next.name);
      next = next.parentNamespace;
    ***REMOVED***
    parts.pop(); // Remove '__rootNamespace__'
    return parts.reverse().join(':');
  ***REMOVED***

  _getFullName() {
    let path = this.path;
    path = (path && path.split(':')) || [];
    path.push(this.name);
    return path.join(':');
  ***REMOVED***

  isRootNamespace() {
    return !this.parentNamespace;
  ***REMOVED***
***REMOVED***

class RootNamespace extends Namespace {
  constructor() {
    super(ROOT_NAMESPACE_NAME, null);
    Namespace.ROOT_NAMESPACE = this;
  ***REMOVED***
***REMOVED***

module.exports.Namespace = Namespace;
module.exports.RootNamespace = RootNamespace;
