// jake/lib/program.js
/*
 * Jake JavaScript build tool
 * Copyright 2112 Matthew Eernisse (mde@fleegix.org)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
*/

let fs = require('fs');
let parseargs = require('./parseargs');
let utils = require('./utils');
let Program;
let usage = require('fs').readFileSync(`${__dirname***REMOVED***/../usage.txt`).toString();
let { Task ***REMOVED*** = require('./task/task');

function die(msg) {
  console.log(msg);
  process.stdout.write('', function () {
    process.stderr.write('', function () {
      process.exit();
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***

let preempts = {
  version: function () {
    die(jake.version);
  ***REMOVED***,
  help: function () {
    die(usage);
  ***REMOVED***
***REMOVED***;

let AVAILABLE_OPTS = [
  { full: 'jakefile',
    abbr: 'f',
    expectValue: true
  ***REMOVED***,
  { full: 'quiet',
    abbr: 'q',
    expectValue: false
  ***REMOVED***,
  { full: 'directory',
    abbr: 'C',
    expectValue: true
  ***REMOVED***,
  { full: 'always-make',
    abbr: 'B',
    expectValue: false
  ***REMOVED***,
  { full: 'tasks',
    abbr: 'T',
    expectValue: false,
    allowValue: true
  ***REMOVED***,
  // Alias t
  { full: 'tasks',
    abbr: 't',
    expectValue: false,
    allowValue: true
  ***REMOVED***,
  // Alias ls
  { full: 'tasks',
    abbr: 'ls',
    expectValue: false,
    allowValue: true
  ***REMOVED***,
  { full: 'help',
    abbr: 'h',
  ***REMOVED***,
  { full: 'version',
    abbr: 'V',
  ***REMOVED***,
  // Alias lowercase v
  { full: 'version',
    abbr: 'v',
  ***REMOVED***,
  { full: 'jakelibdir',
    abbr: 'J',
    expectValue: true
  ***REMOVED***,
  { full: 'allow-rejection',
    abbr: 'ar',
    expectValue: false
  ***REMOVED***
];

Program = function () {
  this.availableOpts = AVAILABLE_OPTS;
  this.opts = {***REMOVED***;
  this.taskNames = null;
  this.taskArgs = null;
  this.envVars = null;
  this.die = die;
***REMOVED***;

Program.prototype = new (function () {

  this.handleErr = function (err) {
    if (jake.listeners('error').length !== 0) {
      jake.emit('error', err);
      return;
    ***REMOVED***

    if (jake.listeners('error').length) {
      jake.emit('error', err);
      return;
    ***REMOVED***

    utils.logger.error('jake aborted.');
    if (err.stack) {
      utils.logger.error(err.stack);
    ***REMOVED***
    else {
      utils.logger.error(err.message);
    ***REMOVED***

    process.stdout.write('', function () {
      process.stderr.write('', function () {
        jake.errorCode = jake.errorCode || 1;
        process.exit(jake.errorCode);
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***;

  this.parseArgs = function (args) {
    let result = (new parseargs.Parser(this.availableOpts)).parse(args);
    this.setOpts(result.opts);
    this.setTaskNames(result.taskNames);
    this.setEnvVars(result.envVars);
  ***REMOVED***;

  this.setOpts = function (options) {
    let opts = options || {***REMOVED***;
    Object.assign(this.opts, opts);
  ***REMOVED***;

  this.internalOpts = function (options) {
    this.availableOpts = this.availableOpts.concat(options);
  ***REMOVED***;

  this.autocompletions = function (cur) {
    let p; let i; let task;
    let commonPrefix = '';
    let matches = [];

    for (p in jake.Task) {
      task = jake.Task[p];
      if (
        'fullName' in task
          && (
            // if empty string, program converts to true
            cur === true ||
            task.fullName.indexOf(cur) === 0
          )
      ) {
        if (matches.length === 0) {
          commonPrefix = task.fullName;
        ***REMOVED***
        else {
          for (i = commonPrefix.length; i > -1; --i) {
            commonPrefix = commonPrefix.substr(0, i);
            if (task.fullName.indexOf(commonPrefix) === 0) {
              break;
            ***REMOVED***
          ***REMOVED***
        ***REMOVED***
        matches.push(task.fullName);
      ***REMOVED***
    ***REMOVED***

    if (matches.length > 1 && commonPrefix === cur) {
      matches.unshift('yes-space');
    ***REMOVED***
    else {
      matches.unshift('no-space');
    ***REMOVED***

    process.stdout.write(matches.join(' '));
  ***REMOVED***;

  this.setTaskNames = function (names) {
    if (names && !Array.isArray(names)) {
      throw new Error('Task names must be an array');
    ***REMOVED***
    this.taskNames = (names && names.length) ? names : ['default'];
  ***REMOVED***;

  this.setEnvVars = function (vars) {
    this.envVars = vars || null;
  ***REMOVED***;

  this.firstPreemptiveOption = function () {
    let opts = this.opts;
    for (let p in opts) {
      if (preempts[p]) {
        return preempts[p];
      ***REMOVED***
    ***REMOVED***
    return false;
  ***REMOVED***;

  this.init = function (configuration) {
    let self = this;
    let config = configuration || {***REMOVED***;
    if (config.options) {
      this.setOpts(config.options);
    ***REMOVED***
    if (config.taskNames) {
      this.setTaskNames(config.taskNames);
    ***REMOVED***
    if (config.envVars) {
      this.setEnvVars(config.envVars);
    ***REMOVED***
    process.addListener('uncaughtException', function (err) {
      self.handleErr(err);
    ***REMOVED***);
    if (!this.opts['allow-rejection']) {
      process.addListener('unhandledRejection', (reason, promise) => {
        utils.logger.error('Unhandled rejection at:', promise, 'reason:', reason);
        self.handleErr(reason);
      ***REMOVED***);
    ***REMOVED***
    if (this.envVars) {
      Object.assign(process.env, this.envVars);
    ***REMOVED***
  ***REMOVED***;

  this.run = function () {
    let rootTask;
    let taskNames;
    let dirname;
    let opts = this.opts;

    if (opts.autocomplete) {
      return this.autocompletions(opts['autocomplete-cur'], opts['autocomplete-prev']);
    ***REMOVED***
    // Run with `jake -T`, just show descriptions
    if (opts.tasks) {
      return jake.showAllTaskDescriptions(opts.tasks);
    ***REMOVED***

    taskNames = this.taskNames;
    if (!(Array.isArray(taskNames) && taskNames.length)) {
      throw new Error('Please pass jake.runTasks an array of task-names');
    ***REMOVED***

    // Set working dir
    dirname = opts.directory;
    if (dirname) {
      if (fs.existsSync(dirname) &&
        fs.statSync(dirname).isDirectory()) {
        process.chdir(dirname);
      ***REMOVED***
      else {
        throw new Error(dirname + ' is not a valid directory path');
      ***REMOVED***
    ***REMOVED***

    rootTask = task(Task.ROOT_TASK_NAME, taskNames, function () {***REMOVED***);
    rootTask._internal = true;

    rootTask.once('complete', function () {
      jake.emit('complete');
    ***REMOVED***);
    jake.emit('start');
    rootTask.invoke();
  ***REMOVED***;

***REMOVED***)();

module.exports.Program = Program;
