// @webassemblyjs/ast/esm/transform/wast-identifier-to-index/index.js
function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; ***REMOVED***; ***REMOVED*** else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; ***REMOVED***; ***REMOVED*** return _typeof(obj); ***REMOVED***

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); ***REMOVED***

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); ***REMOVED***

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); ***REMOVED***

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; ***REMOVED*** return arr2; ***REMOVED***

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; ***REMOVED*** ***REMOVED*** catch (err) { _d = true; _e = err; ***REMOVED*** finally { try { if (!_n && _i["return"] != null) _i["return"](); ***REMOVED*** finally { if (_d) throw _e; ***REMOVED*** ***REMOVED*** return _arr; ***REMOVED***

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; ***REMOVED***

import { isBlock, isFunc, isIdentifier, numberLiteralFromRaw, traverse ***REMOVED*** from "../../index";
import { moduleContextFromModuleAST ***REMOVED*** from "../ast-module-to-module-context"; // FIXME(sven): do the same with all block instructions, must be more generic here

function newUnexpectedFunction(i) {
  return new Error("unknown function at offset: " + i);
***REMOVED***

export function transform(ast) {
  var module = null;
  traverse(ast, {
    Module: function (_Module) {
      function Module(_x) {
        return _Module.apply(this, arguments);
      ***REMOVED***

      Module.toString = function () {
        return _Module.toString();
      ***REMOVED***;

      return Module;
    ***REMOVED***(function (path) {
      module = path.node;
    ***REMOVED***)
  ***REMOVED***);

  if (module == null) {
    throw new Error("Module not foudn in program");
  ***REMOVED***

  var moduleContext = moduleContextFromModuleAST(module); // Transform the actual instruction in function bodies

  traverse(ast, {
    Func: function (_Func) {
      function Func(_x2) {
        return _Func.apply(this, arguments);
      ***REMOVED***

      Func.toString = function () {
        return _Func.toString();
      ***REMOVED***;

      return Func;
    ***REMOVED***(function (path) {
      transformFuncPath(path, moduleContext);
    ***REMOVED***),
    Start: function (_Start) {
      function Start(_x3) {
        return _Start.apply(this, arguments);
      ***REMOVED***

      Start.toString = function () {
        return _Start.toString();
      ***REMOVED***;

      return Start;
    ***REMOVED***(function (path) {
      var index = path.node.index;

      if (isIdentifier(index) === true) {
        var offsetInModule = moduleContext.getFunctionOffsetByIdentifier(index.value);

        if (typeof offsetInModule === "undefined") {
          throw newUnexpectedFunction(index.value);
        ***REMOVED*** // Replace the index Identifier
        // $FlowIgnore: reference?


        path.node.index = numberLiteralFromRaw(offsetInModule);
      ***REMOVED***
    ***REMOVED***)
  ***REMOVED***);
***REMOVED***

function transformFuncPath(funcPath, moduleContext) {
  var funcNode = funcPath.node;
  var signature = funcNode.signature;

  if (signature.type !== "Signature") {
    throw new Error("Function signatures must be denormalised before execution");
  ***REMOVED***

  var params = signature.params; // Add func locals in the context

  params.forEach(function (p) {
    return moduleContext.addLocal(p.valtype);
  ***REMOVED***);
  traverse(funcNode, {
    Instr: function (_Instr) {
      function Instr(_x4) {
        return _Instr.apply(this, arguments);
      ***REMOVED***

      Instr.toString = function () {
        return _Instr.toString();
      ***REMOVED***;

      return Instr;
    ***REMOVED***(function (instrPath) {
      var instrNode = instrPath.node;
      /**
       * Local access
       */

      if (instrNode.id === "get_local" || instrNode.id === "set_local" || instrNode.id === "tee_local") {
        var _instrNode$args = _slicedToArray(instrNode.args, 1),
            firstArg = _instrNode$args[0];

        if (firstArg.type === "Identifier") {
          var offsetInParams = params.findIndex(function (_ref) {
            var id = _ref.id;
            return id === firstArg.value;
          ***REMOVED***);

          if (offsetInParams === -1) {
            throw new Error("".concat(firstArg.value, " not found in ").concat(instrNode.id, ": not declared in func params"));
          ***REMOVED*** // Replace the Identifer node by our new NumberLiteral node


          instrNode.args[0] = numberLiteralFromRaw(offsetInParams);
        ***REMOVED***
      ***REMOVED***
      /**
       * Global access
       */


      if (instrNode.id === "get_global" || instrNode.id === "set_global") {
        var _instrNode$args2 = _slicedToArray(instrNode.args, 1),
            _firstArg = _instrNode$args2[0];

        if (isIdentifier(_firstArg) === true) {
          var globalOffset = moduleContext.getGlobalOffsetByIdentifier( // $FlowIgnore: reference?
          _firstArg.value);

          if (typeof globalOffset === "undefined") {
            // $FlowIgnore: reference?
            throw new Error("global ".concat(_firstArg.value, " not found in module"));
          ***REMOVED*** // Replace the Identifer node by our new NumberLiteral node


          instrNode.args[0] = numberLiteralFromRaw(globalOffset);
        ***REMOVED***
      ***REMOVED***
      /**
       * Labels lookup
       */


      if (instrNode.id === "br") {
        var _instrNode$args3 = _slicedToArray(instrNode.args, 1),
            _firstArg2 = _instrNode$args3[0];

        if (isIdentifier(_firstArg2) === true) {
          // if the labels is not found it is going to be replaced with -1
          // which is invalid.
          var relativeBlockCount = -1; // $FlowIgnore: reference?

          instrPath.findParent(function (_ref2) {
            var node = _ref2.node;

            if (isBlock(node)) {
              relativeBlockCount++; // $FlowIgnore: reference?

              var name = node.label || node.name;

              if (_typeof(name) === "object") {
                // $FlowIgnore: isIdentifier ensures that
                if (name.value === _firstArg2.value) {
                  // Found it
                  return false;
                ***REMOVED***
              ***REMOVED***
            ***REMOVED***

            if (isFunc(node)) {
              return false;
            ***REMOVED***
          ***REMOVED***); // Replace the Identifer node by our new NumberLiteral node

          instrNode.args[0] = numberLiteralFromRaw(relativeBlockCount);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***),

    /**
     * Func lookup
     */
    CallInstruction: function (_CallInstruction) {
      function CallInstruction(_x5) {
        return _CallInstruction.apply(this, arguments);
      ***REMOVED***

      CallInstruction.toString = function () {
        return _CallInstruction.toString();
      ***REMOVED***;

      return CallInstruction;
    ***REMOVED***(function (_ref3) {
      var node = _ref3.node;
      var index = node.index;

      if (isIdentifier(index) === true) {
        var offsetInModule = moduleContext.getFunctionOffsetByIdentifier(index.value);

        if (typeof offsetInModule === "undefined") {
          throw newUnexpectedFunction(index.value);
        ***REMOVED*** // Replace the index Identifier
        // $FlowIgnore: reference?


        node.index = numberLiteralFromRaw(offsetInModule);
      ***REMOVED***
    ***REMOVED***)
  ***REMOVED***);
***REMOVED***
