// spdy-transport/lib/spdy-transport/connection.js
'use strict'

var util = require('util')
var transport = require('../spdy-transport')

var debug = {
  server: require('debug')('spdy:connection:server'),
  client: require('debug')('spdy:connection:client')
***REMOVED***
var EventEmitter = require('events').EventEmitter

var Stream = transport.Stream

function Connection (socket, options) {
  EventEmitter.call(this)

  var state = {***REMOVED***
  this._spdyState = state

  // NOTE: There's a big trick here. Connection is used as a `this` argument
  // to the wrapped `connection` event listener.
  // socket end doesn't necessarly mean connection drop
  this.httpAllowHalfOpen = true

  state.timeout = new transport.utils.Timeout(this)

  // Protocol info
  state.protocol = transport.protocol[options.protocol]
  state.version = null
  state.constants = state.protocol.constants
  state.pair = null
  state.isServer = options.isServer

  // Root of priority tree (i.e. stream id = 0)
  state.priorityRoot = new transport.Priority({
    defaultWeight: state.constants.DEFAULT_WEIGHT,
    maxCount: transport.protocol.base.constants.MAX_PRIORITY_STREAMS
  ***REMOVED***)

  // Defaults
  state.maxStreams = options.maxStreams ||
                     state.constants.MAX_CONCURRENT_STREAMS

  state.autoSpdy31 = options.protocol.name !== 'h2' && options.autoSpdy31
  state.acceptPush = options.acceptPush === undefined
    ? !state.isServer
    : options.acceptPush

  if (options.maxChunk === false) { state.maxChunk = Infinity ***REMOVED*** else if (options.maxChunk === undefined) { state.maxChunk = transport.protocol.base.constants.DEFAULT_MAX_CHUNK ***REMOVED*** else {
    state.maxChunk = options.maxChunk
  ***REMOVED***

  // Connection-level flow control
  var windowSize = options.windowSize || 1 << 20
  state.window = new transport.Window({
    id: 0,
    isServer: state.isServer,
    recv: {
      size: state.constants.DEFAULT_WINDOW,
      max: state.constants.MAX_INITIAL_WINDOW_SIZE
    ***REMOVED***,
    send: {
      size: state.constants.DEFAULT_WINDOW,
      max: state.constants.MAX_INITIAL_WINDOW_SIZE
    ***REMOVED***
  ***REMOVED***)

  // It starts with DEFAULT_WINDOW, update must be sent to change it on client
  state.window.recv.setMax(windowSize)

  // Boilerplate for Stream constructor
  state.streamWindow = new transport.Window({
    id: -1,
    isServer: state.isServer,
    recv: {
      size: windowSize,
      max: state.constants.MAX_INITIAL_WINDOW_SIZE
    ***REMOVED***,
    send: {
      size: state.constants.DEFAULT_WINDOW,
      max: state.constants.MAX_INITIAL_WINDOW_SIZE
    ***REMOVED***
  ***REMOVED***)

  // Various state info
  state.pool = state.protocol.compressionPool.create(options.headerCompression)
  state.counters = {
    push: 0,
    stream: 0
  ***REMOVED***

  // Init streams list
  state.stream = {
    map: {***REMOVED***,
    count: 0,
    nextId: state.isServer ? 2 : 1,
    lastId: {
      both: 0,
      received: 0
    ***REMOVED***
  ***REMOVED***
  state.ping = {
    nextId: state.isServer ? 2 : 1,
    map: {***REMOVED***
  ***REMOVED***
  state.goaway = false

  // Debug
  state.debug = state.isServer ? debug.server : debug.client

  // X-Forwarded feature
  state.xForward = null

  // Create parser and hole for framer
  state.parser = state.protocol.parser.create({
    // NOTE: needed to distinguish ping from ping ACK in SPDY
    isServer: state.isServer,
    window: state.window
  ***REMOVED***)
  state.framer = state.protocol.framer.create({
    window: state.window,
    timeout: state.timeout
  ***REMOVED***)

  // SPDY has PUSH enabled on servers
  if (state.protocol.name === 'spdy') {
    state.framer.enablePush(state.isServer)
  ***REMOVED***

  if (!state.isServer) { state.parser.skipPreface() ***REMOVED***

  this.socket = socket

  this._init()
***REMOVED***
util.inherits(Connection, EventEmitter)
exports.Connection = Connection

Connection.create = function create (socket, options) {
  return new Connection(socket, options)
***REMOVED***

Connection.prototype._init = function init () {
  var self = this
  var state = this._spdyState
  var pool = state.pool

  // Initialize session window
  state.window.recv.on('drain', function () {
    self._onSessionWindowDrain()
  ***REMOVED***)

  // Initialize parser
  state.parser.on('data', function (frame) {
    self._handleFrame(frame)
  ***REMOVED***)
  state.parser.once('version', function (version) {
    self._onVersion(version)
  ***REMOVED***)

  // Propagate parser errors
  state.parser.on('error', function (err) {
    self._onParserError(err)
  ***REMOVED***)

  // Propagate framer errors
  state.framer.on('error', function (err) {
    self.emit('error', err)
  ***REMOVED***)

  this.socket.pipe(state.parser)
  state.framer.pipe(this.socket)

  // Allow high-level api to catch socket errors
  this.socket.on('error', function onSocketError (e) {
    self.emit('error', e)
  ***REMOVED***)

  this.socket.once('close', function onclose (hadError) {
    var err
    if (hadError) {
      err = new Error('socket hang up')
      err.code = 'ECONNRESET'
    ***REMOVED***

    self.destroyStreams(err)
    self.emit('close')

    if (state.pair) {
      pool.put(state.pair)
    ***REMOVED***

    state.framer.resume()
  ***REMOVED***)

  // Reset timeout on close
  this.once('close', function () {
    self.setTimeout(0)
  ***REMOVED***)

  function _onWindowOverflow () {
    self._onWindowOverflow()
  ***REMOVED***

  state.window.recv.on('overflow', _onWindowOverflow)
  state.window.send.on('overflow', _onWindowOverflow)

  // Do not allow half-open connections
  this.socket.allowHalfOpen = false
***REMOVED***

Connection.prototype._onVersion = function _onVersion (version) {
  var state = this._spdyState
  var prev = state.version
  var parser = state.parser
  var framer = state.framer
  var pool = state.pool

  state.version = version
  state.debug('id=0 version=%d', version)

  // Ignore transition to 3.1
  if (!prev) {
    state.pair = pool.get(version)
    parser.setCompression(state.pair)
    framer.setCompression(state.pair)
  ***REMOVED***
  framer.setVersion(version)

  if (!state.isServer) {
    framer.prefaceFrame()
    if (state.xForward !== null) {
      framer.xForwardedFor({ host: state.xForward ***REMOVED***)
    ***REMOVED***
  ***REMOVED***

  // Send preface+settings frame (once)
  framer.settingsFrame({
    max_header_list_size: state.constants.DEFAULT_MAX_HEADER_LIST_SIZE,
    max_concurrent_streams: state.maxStreams,
    enable_push: state.acceptPush ? 1 : 0,
    initial_window_size: state.window.recv.max
  ***REMOVED***)

  // Update session window
  if (state.version >= 3.1 || (state.isServer && state.autoSpdy31)) { this._onSessionWindowDrain() ***REMOVED***

  this.emit('version', version)
***REMOVED***

Connection.prototype._onParserError = function _onParserError (err) {
  var state = this._spdyState

  // Prevent further errors
  state.parser.kill()

  // Send GOAWAY
  if (err instanceof transport.protocol.base.utils.ProtocolError) {
    this._goaway({
      lastId: state.stream.lastId.both,
      code: err.code,
      extra: err.message,
      send: true
    ***REMOVED***)
  ***REMOVED***

  this.emit('error', err)
***REMOVED***

Connection.prototype._handleFrame = function _handleFrame (frame) {
  var state = this._spdyState

  state.debug('id=0 frame', frame)
  state.timeout.reset()

  // For testing purposes
  this.emit('frame', frame)

  var stream

  // Session window update
  if (frame.type === 'WINDOW_UPDATE' && frame.id === 0) {
    if (state.version < 3.1 && state.autoSpdy31) {
      state.debug('id=0 switch version to 3.1')
      state.version = 3.1
      this.emit('version', 3.1)
    ***REMOVED***
    state.window.send.update(frame.delta)
    return
  ***REMOVED***

  if (state.isServer && frame.type === 'PUSH_PROMISE') {
    state.debug('id=0 server PUSH_PROMISE')
    this._goaway({
      lastId: state.stream.lastId.both,
      code: 'PROTOCOL_ERROR',
      send: true
    ***REMOVED***)
    return
  ***REMOVED***

  if (!stream && frame.id !== undefined) {
    // Load created one
    stream = state.stream.map[frame.id]

    // Fail if not found
    if (!stream &&
        frame.type !== 'HEADERS' &&
        frame.type !== 'PRIORITY' &&
        frame.type !== 'RST') {
      // Other side should destroy the stream upon receiving GOAWAY
      if (this._isGoaway(frame.id)) { return ***REMOVED***

      state.debug('id=0 stream=%d not found', frame.id)
      state.framer.rstFrame({ id: frame.id, code: 'INVALID_STREAM' ***REMOVED***)
      return
    ***REMOVED***
  ***REMOVED***

  // Create new stream
  if (!stream && frame.type === 'HEADERS') {
    this._handleHeaders(frame)
    return
  ***REMOVED***

  if (stream) {
    stream._handleFrame(frame)
  ***REMOVED*** else if (frame.type === 'SETTINGS') {
    this._handleSettings(frame.settings)
  ***REMOVED*** else if (frame.type === 'ACK_SETTINGS') {
    // TODO(indutny): handle it one day
  ***REMOVED*** else if (frame.type === 'PING') {
    this._handlePing(frame)
  ***REMOVED*** else if (frame.type === 'GOAWAY') {
    this._handleGoaway(frame)
  ***REMOVED*** else if (frame.type === 'X_FORWARDED_FOR') {
    // Set X-Forwarded-For only once
    if (state.xForward === null) {
      state.xForward = frame.host
    ***REMOVED***
  ***REMOVED*** else if (frame.type === 'PRIORITY') {
    // TODO(indutny): handle this
  ***REMOVED*** else {
    state.debug('id=0 unknown frame type: %s', frame.type)
  ***REMOVED***
***REMOVED***

Connection.prototype._onWindowOverflow = function _onWindowOverflow () {
  var state = this._spdyState
  state.debug('id=0 window overflow')
  this._goaway({
    lastId: state.stream.lastId.both,
    code: 'FLOW_CONTROL_ERROR',
    send: true
  ***REMOVED***)
***REMOVED***

Connection.prototype._isGoaway = function _isGoaway (id) {
  var state = this._spdyState
  if (state.goaway !== false && state.goaway < id) { return true ***REMOVED***
  return false
***REMOVED***

Connection.prototype._getId = function _getId () {
  var state = this._spdyState

  var id = state.stream.nextId
  state.stream.nextId += 2
  return id
***REMOVED***

Connection.prototype._createStream = function _createStream (uri) {
  var state = this._spdyState
  var id = uri.id
  if (id === undefined) { id = this._getId() ***REMOVED***

  var isGoaway = this._isGoaway(id)

  if (uri.push && !state.acceptPush) {
    state.debug('id=0 push disabled promisedId=%d', id)

    // Fatal error
    this._goaway({
      lastId: state.stream.lastId.both,
      code: 'PROTOCOL_ERROR',
      send: true
    ***REMOVED***)
    isGoaway = true
  ***REMOVED***

  var stream = new Stream(this, {
    id: id,
    request: uri.request !== false,
    method: uri.method,
    path: uri.path,
    host: uri.host,
    priority: uri.priority,
    headers: uri.headers,
    parent: uri.parent,
    readable: !isGoaway && uri.readable,
    writable: !isGoaway && uri.writable
  ***REMOVED***)
  var self = this

  // Just an empty stream for API consistency
  if (isGoaway) {
    return stream
  ***REMOVED***

  state.stream.lastId.both = Math.max(state.stream.lastId.both, id)

  state.debug('id=0 add stream=%d', stream.id)
  state.stream.map[stream.id] = stream
  state.stream.count++
  state.counters.stream++
  if (stream.parent !== null) {
    state.counters.push++
  ***REMOVED***

  stream.once('close', function () {
    self._removeStream(stream)
  ***REMOVED***)

  return stream
***REMOVED***

Connection.prototype._handleHeaders = function _handleHeaders (frame) {
  var state = this._spdyState

  // Must be HEADERS frame after stream close
  if (frame.id <= state.stream.lastId.received) { return ***REMOVED***

  // Someone is using our ids!
  if ((frame.id + state.stream.nextId) % 2 === 0) {
    state.framer.rstFrame({ id: frame.id, code: 'PROTOCOL_ERROR' ***REMOVED***)
    return
  ***REMOVED***

  var stream = this._createStream({
    id: frame.id,
    request: false,
    method: frame.headers[':method'],
    path: frame.headers[':path'],
    host: frame.headers[':authority'],
    priority: frame.priority,
    headers: frame.headers,
    writable: frame.writable
  ***REMOVED***)

  // GOAWAY
  if (this._isGoaway(stream.id)) {
    return
  ***REMOVED***

  state.stream.lastId.received = Math.max(
    state.stream.lastId.received,
    stream.id
  )

  // TODO(indutny) handle stream limit
  if (!this.emit('stream', stream)) {
    // No listeners was set - abort the stream
    stream.abort()
    return
  ***REMOVED***

  // Create fake frame to simulate end of the data
  if (frame.fin) {
    stream._handleFrame({ type: 'FIN', fin: true ***REMOVED***)
  ***REMOVED***

  return stream
***REMOVED***

Connection.prototype._onSessionWindowDrain = function _onSessionWindowDrain () {
  var state = this._spdyState
  if (state.version < 3.1 && !(state.isServer && state.autoSpdy31)) {
    return
  ***REMOVED***

  var delta = state.window.recv.getDelta()
  if (delta === 0) {
    return
  ***REMOVED***

  state.debug('id=0 session window drain, update by %d', delta)

  state.framer.windowUpdateFrame({
    id: 0,
    delta: delta
  ***REMOVED***)
  state.window.recv.update(delta)
***REMOVED***

Connection.prototype.start = function start (version) {
  this._spdyState.parser.setVersion(version)
***REMOVED***

// Mostly for testing
Connection.prototype.getVersion = function getVersion () {
  return this._spdyState.version
***REMOVED***

Connection.prototype._handleSettings = function _handleSettings (settings) {
  var state = this._spdyState

  state.framer.ackSettingsFrame()

  this._setDefaultWindow(settings)
  if (settings.max_frame_size) { state.framer.setMaxFrameSize(settings.max_frame_size) ***REMOVED***

  // TODO(indutny): handle max_header_list_size
  if (settings.header_table_size) {
    try {
      state.pair.compress.updateTableSize(settings.header_table_size)
    ***REMOVED*** catch (e) {
      this._goaway({
        lastId: 0,
        code: 'PROTOCOL_ERROR',
        send: true
      ***REMOVED***)
      return
    ***REMOVED***
  ***REMOVED***

  // HTTP2 clients needs to enable PUSH streams explicitly
  if (state.protocol.name !== 'spdy') {
    if (settings.enable_push === undefined) {
      state.framer.enablePush(state.isServer)
    ***REMOVED*** else {
      state.framer.enablePush(settings.enable_push === 1)
    ***REMOVED***
  ***REMOVED***

  // TODO(indutny): handle max_concurrent_streams
***REMOVED***

Connection.prototype._setDefaultWindow = function _setDefaultWindow (settings) {
  if (settings.initial_window_size === undefined) {
    return
  ***REMOVED***

  var state = this._spdyState

  // Update defaults
  var window = state.streamWindow
  window.send.setMax(settings.initial_window_size)

  // Update existing streams
  Object.keys(state.stream.map).forEach(function (id) {
    var stream = state.stream.map[id]
    var window = stream._spdyState.window

    window.send.updateMax(settings.initial_window_size)
  ***REMOVED***)
***REMOVED***

Connection.prototype._handlePing = function handlePing (frame) {
  var self = this
  var state = this._spdyState

  // Handle incoming PING
  if (!frame.ack) {
    state.framer.pingFrame({
      opaque: frame.opaque,
      ack: true
    ***REMOVED***)

    self.emit('ping', frame.opaque)
    return
  ***REMOVED***

  // Handle reply PING
  var hex = frame.opaque.toString('hex')
  if (!state.ping.map[hex]) {
    return
  ***REMOVED***
  var ping = state.ping.map[hex]
  delete state.ping.map[hex]

  if (ping.cb) {
    ping.cb(null)
  ***REMOVED***
***REMOVED***

Connection.prototype._handleGoaway = function handleGoaway (frame) {
  this._goaway({
    lastId: frame.lastId,
    code: frame.code,
    send: false
  ***REMOVED***)
***REMOVED***

Connection.prototype.ping = function ping (callback) {
  var state = this._spdyState

  // HTTP2 is using 8-byte opaque
  var opaque = Buffer.alloc(state.constants.PING_OPAQUE_SIZE)
  opaque.fill(0)
  opaque.writeUInt32BE(state.ping.nextId, opaque.length - 4)
  state.ping.nextId += 2

  state.ping.map[opaque.toString('hex')] = { cb: callback ***REMOVED***
  state.framer.pingFrame({
    opaque: opaque,
    ack: false
  ***REMOVED***)
***REMOVED***

Connection.prototype.getCounter = function getCounter (name) {
  return this._spdyState.counters[name]
***REMOVED***

Connection.prototype.reserveStream = function reserveStream (uri, callback) {
  var stream = this._createStream(uri)

  // GOAWAY
  if (this._isGoaway(stream.id)) {
    var err = new Error('Can\'t send request after GOAWAY')
    process.nextTick(function () {
      if (callback) { callback(err) ***REMOVED*** else {
        stream.emit('error', err)
      ***REMOVED***
    ***REMOVED***)
    return stream
  ***REMOVED***

  if (callback) {
    process.nextTick(function () {
      callback(null, stream)
    ***REMOVED***)
  ***REMOVED***

  return stream
***REMOVED***

Connection.prototype.request = function request (uri, callback) {
  var stream = this.reserveStream(uri, function (err) {
    if (err) {
      if (callback) {
        callback(err)
      ***REMOVED*** else {
        stream.emit('error', err)
      ***REMOVED***
      return
    ***REMOVED***

    if (stream._wasSent()) {
      if (callback) {
        callback(null, stream)
      ***REMOVED***
      return
    ***REMOVED***

    stream.send(function (err) {
      if (err) {
        if (callback) { return callback(err) ***REMOVED*** else { return stream.emit('error', err) ***REMOVED***
      ***REMOVED***

      if (callback) {
        callback(null, stream)
      ***REMOVED***
    ***REMOVED***)
  ***REMOVED***)

  return stream
***REMOVED***

Connection.prototype._removeStream = function _removeStream (stream) {
  var state = this._spdyState

  state.debug('id=0 remove stream=%d', stream.id)
  delete state.stream.map[stream.id]
  state.stream.count--

  if (state.stream.count === 0) {
    this.emit('_streamDrain')
  ***REMOVED***
***REMOVED***

Connection.prototype._goaway = function _goaway (params) {
  var state = this._spdyState
  var self = this

  state.goaway = params.lastId
  state.debug('id=0 goaway from=%d', state.goaway)

  Object.keys(state.stream.map).forEach(function (id) {
    var stream = state.stream.map[id]

    // Abort every stream started after GOAWAY
    if (stream.id <= params.lastId) {
      return
    ***REMOVED***

    stream.abort()
    stream.emit('error', new Error('New stream after GOAWAY'))
  ***REMOVED***)

  function finish () {
    // Destroy socket if there are no streams
    if (state.stream.count === 0 || params.code !== 'OK') {
      // No further frames should be processed
      state.parser.kill()

      process.nextTick(function () {
        var err = new Error('Fatal error: ' + params.code)
        self._onStreamDrain(err)
      ***REMOVED***)
      return
    ***REMOVED***

    self.on('_streamDrain', self._onStreamDrain)
  ***REMOVED***

  if (params.send) {
    // Make sure that GOAWAY frame is sent before dumping framer
    state.framer.goawayFrame({
      lastId: params.lastId,
      code: params.code,
      extra: params.extra
    ***REMOVED***, finish)
  ***REMOVED*** else {
    finish()
  ***REMOVED***
***REMOVED***

Connection.prototype._onStreamDrain = function _onStreamDrain (error) {
  var state = this._spdyState

  state.debug('id=0 _onStreamDrain')

  state.framer.dump()
  state.framer.unpipe(this.socket)
  state.framer.resume()

  if (this.socket.destroySoon) {
    this.socket.destroySoon()
  ***REMOVED***
  this.emit('close', error)
***REMOVED***

Connection.prototype.end = function end (callback) {
  var state = this._spdyState

  if (callback) {
    this.once('close', callback)
  ***REMOVED***
  this._goaway({
    lastId: state.stream.lastId.both,
    code: 'OK',
    send: true
  ***REMOVED***)
***REMOVED***

Connection.prototype.destroyStreams = function destroyStreams (err) {
  var state = this._spdyState
  Object.keys(state.stream.map).forEach(function (id) {
    var stream = state.stream.map[id]

    stream.destroy()
    if (err) {
      stream.emit('error', err)
    ***REMOVED***
  ***REMOVED***)
***REMOVED***

Connection.prototype.isServer = function isServer () {
  return this._spdyState.isServer
***REMOVED***

Connection.prototype.getXForwardedFor = function getXForwardFor () {
  return this._spdyState.xForward
***REMOVED***

Connection.prototype.sendXForwardedFor = function sendXForwardedFor (host) {
  var state = this._spdyState
  if (state.version !== null) {
    state.framer.xForwardedFor({ host: host ***REMOVED***)
  ***REMOVED*** else {
    state.xForward = host
  ***REMOVED***
***REMOVED***

Connection.prototype.pushPromise = function pushPromise (parent, uri, callback) {
  var state = this._spdyState

  var stream = this._createStream({
    request: false,
    parent: parent,
    method: uri.method,
    path: uri.path,
    host: uri.host,
    priority: uri.priority,
    headers: uri.headers,
    readable: false
  ***REMOVED***)

  var err

  // TODO(indutny): deduplicate this logic somehow
  if (this._isGoaway(stream.id)) {
    err = new Error('Can\'t send PUSH_PROMISE after GOAWAY')

    process.nextTick(function () {
      if (callback) {
        callback(err)
      ***REMOVED*** else {
        stream.emit('error', err)
      ***REMOVED***
    ***REMOVED***)
    return stream
  ***REMOVED***

  if (uri.push && !state.acceptPush) {
    err = new Error(
      'Can\'t send PUSH_PROMISE, other side won\'t accept it')
    process.nextTick(function () {
      if (callback) { callback(err) ***REMOVED*** else {
        stream.emit('error', err)
      ***REMOVED***
    ***REMOVED***)
    return stream
  ***REMOVED***

  stream._sendPush(uri.status, uri.response, function (err) {
    if (!callback) {
      if (err) {
        stream.emit('error', err)
      ***REMOVED***
      return
    ***REMOVED***

    if (err) { return callback(err) ***REMOVED***
    callback(null, stream)
  ***REMOVED***)

  return stream
***REMOVED***

Connection.prototype.setTimeout = function setTimeout (delay, callback) {
  var state = this._spdyState

  state.timeout.set(delay, callback)
***REMOVED***
