// @typescript-eslint/typescript-estree/dist/create-program/createIsolatedProgram.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; ***REMOVED*** ***REMOVED***;
    ***REMOVED***
    Object.defineProperty(o, k2, desc);
***REMOVED***) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
***REMOVED***));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v ***REMOVED***);
***REMOVED***) : function(o, v) {
    o["default"] = v;
***REMOVED***);
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {***REMOVED***;
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
***REMOVED***;
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod ***REMOVED***;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.createIsolatedProgram = void 0;
const debug_1 = __importDefault(require("debug"));
const ts = __importStar(require("typescript"));
const getScriptKind_1 = require("./getScriptKind");
const shared_1 = require("./shared");
const log = (0, debug_1.default)('typescript-eslint:typescript-estree:createIsolatedProgram');
/**
 * @param code The code of the file being linted
 * @returns Returns a new source file and program corresponding to the linted code
 */
function createIsolatedProgram(parseSettings) {
    log('Getting isolated program in %s mode for: %s', parseSettings.jsx ? 'TSX' : 'TS', parseSettings.filePath);
    const compilerHost = {
        fileExists() {
            return true;
        ***REMOVED***,
        getCanonicalFileName() {
            return parseSettings.filePath;
        ***REMOVED***,
        getCurrentDirectory() {
            return '';
        ***REMOVED***,
        getDirectories() {
            return [];
        ***REMOVED***,
        getDefaultLibFileName() {
            return 'lib.d.ts';
        ***REMOVED***,
        // TODO: Support Windows CRLF
        getNewLine() {
            return '\n';
        ***REMOVED***,
        getSourceFile(filename) {
            return ts.createSourceFile(filename, parseSettings.code, ts.ScriptTarget.Latest, 
            /* setParentNodes */ true, (0, getScriptKind_1.getScriptKind)(parseSettings.filePath, parseSettings.jsx));
        ***REMOVED***,
        readFile() {
            return undefined;
        ***REMOVED***,
        useCaseSensitiveFileNames() {
            return true;
        ***REMOVED***,
        writeFile() {
            return null;
        ***REMOVED***,
    ***REMOVED***;
    const program = ts.createProgram([parseSettings.filePath], Object.assign({ noResolve: true, target: ts.ScriptTarget.Latest, jsx: parseSettings.jsx ? ts.JsxEmit.Preserve : undefined ***REMOVED***, (0, shared_1.createDefaultCompilerOptionsFromExtra)(parseSettings)), compilerHost);
    const ast = program.getSourceFile(parseSettings.filePath);
    if (!ast) {
        throw new Error('Expected an ast to be returned for the single-file isolated program.');
    ***REMOVED***
    return { ast, program ***REMOVED***;
***REMOVED***
exports.createIsolatedProgram = createIsolatedProgram;
