// webpack-sources/lib/CompatSource.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

const Source = require("./Source");

class CompatSource extends Source {
	static from(sourceLike) {
		return sourceLike instanceof Source
			? sourceLike
			: new CompatSource(sourceLike);
	***REMOVED***

	constructor(sourceLike) {
		super();
		this._sourceLike = sourceLike;
	***REMOVED***

	source() {
		return this._sourceLike.source();
	***REMOVED***

	buffer() {
		if (typeof this._sourceLike.buffer === "function") {
			return this._sourceLike.buffer();
		***REMOVED***
		return super.buffer();
	***REMOVED***

	size() {
		if (typeof this._sourceLike.size === "function") {
			return this._sourceLike.size();
		***REMOVED***
		return super.size();
	***REMOVED***

	map(options) {
		if (typeof this._sourceLike.map === "function") {
			return this._sourceLike.map(options);
		***REMOVED***
		return super.map(options);
	***REMOVED***

	sourceAndMap(options) {
		if (typeof this._sourceLike.sourceAndMap === "function") {
			return this._sourceLike.sourceAndMap(options);
		***REMOVED***
		return super.sourceAndMap(options);
	***REMOVED***

	updateHash(hash) {
		if (typeof this._sourceLike.updateHash === "function") {
			return this._sourceLike.updateHash(hash);
		***REMOVED***
		if (typeof this._sourceLike.map === "function") {
			throw new Error(
				"A Source-like object with a 'map' method must also provide an 'updateHash' method"
			);
		***REMOVED***
		hash.update(this.buffer());
	***REMOVED***
***REMOVED***

module.exports = CompatSource;
