// webpack-sources/lib/OriginalSource.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

const Source = require("./Source");
const { SourceNode ***REMOVED*** = require("source-map");
const { SourceListMap ***REMOVED*** = require("source-list-map");
const { getSourceAndMap, getMap ***REMOVED*** = require("./helpers");

const SPLIT_REGEX = /(?!$)[^\n\r;{***REMOVED***]*[\n\r;{***REMOVED***]*/g;

function _splitCode(code) {
	return code.match(SPLIT_REGEX) || [];
***REMOVED***

class OriginalSource extends Source {
	constructor(value, name) {
		super();
		const isBuffer = Buffer.isBuffer(value);
		this._value = isBuffer ? undefined : value;
		this._valueAsBuffer = isBuffer ? value : undefined;
		this._name = name;
	***REMOVED***

	getName() {
		return this._name;
	***REMOVED***

	source() {
		if (this._value === undefined) {
			this._value = this._valueAsBuffer.toString("utf-8");
		***REMOVED***
		return this._value;
	***REMOVED***

	buffer() {
		if (this._valueAsBuffer === undefined) {
			this._valueAsBuffer = Buffer.from(this._value, "utf-8");
		***REMOVED***
		return this._valueAsBuffer;
	***REMOVED***

	map(options) {
		return getMap(this, options);
	***REMOVED***

	sourceAndMap(options) {
		return getSourceAndMap(this, options);
	***REMOVED***

	node(options) {
		if (this._value === undefined) {
			this._value = this._valueAsBuffer.toString("utf-8");
		***REMOVED***
		const value = this._value;
		const name = this._name;
		const lines = value.split("\n");
		const node = new SourceNode(
			null,
			null,
			null,
			lines.map(function (line, idx) {
				let pos = 0;
				if (options && options.columns === false) {
					const content = line + (idx !== lines.length - 1 ? "\n" : "");
					return new SourceNode(idx + 1, 0, name, content);
				***REMOVED***
				return new SourceNode(
					null,
					null,
					null,
					_splitCode(line + (idx !== lines.length - 1 ? "\n" : "")).map(
						function (item) {
							if (/^\s*$/.test(item)) {
								pos += item.length;
								return item;
							***REMOVED***
							const res = new SourceNode(idx + 1, pos, name, item);
							pos += item.length;
							return res;
						***REMOVED***
					)
				);
			***REMOVED***)
		);
		node.setSourceContent(name, value);
		return node;
	***REMOVED***

	listMap(options) {
		if (this._value === undefined) {
			this._value = this._valueAsBuffer.toString("utf-8");
		***REMOVED***
		return new SourceListMap(this._value, this._name, this._value);
	***REMOVED***

	updateHash(hash) {
		if (this._valueAsBuffer === undefined) {
			this._valueAsBuffer = Buffer.from(this._value, "utf-8");
		***REMOVED***
		hash.update("OriginalSource");
		hash.update(this._valueAsBuffer);
		hash.update(this._name || "");
	***REMOVED***
***REMOVED***

module.exports = OriginalSource;
