// ajv/lib/compile/rules.ts
import type {AddedKeywordDefinition***REMOVED*** from "../types"

const _jsonTypes = ["string", "number", "integer", "boolean", "null", "object", "array"] as const

export type JSONType = (typeof _jsonTypes)[number]

const jsonTypes: Set<string> = new Set(_jsonTypes)

export function isJSONType(x: unknown): x is JSONType {
  return typeof x == "string" && jsonTypes.has(x)
***REMOVED***

type ValidationTypes = {
  [K in JSONType]: boolean | RuleGroup | undefined
***REMOVED***

export interface ValidationRules {
  rules: RuleGroup[]
  post: RuleGroup
  all: {[Key in string]?: boolean | Rule***REMOVED*** // rules that have to be validated
  keywords: {[Key in string]?: boolean***REMOVED*** // all known keywords (superset of "all")
  types: ValidationTypes
***REMOVED***

export interface RuleGroup {
  type?: JSONType
  rules: Rule[]
***REMOVED***

// This interface wraps KeywordDefinition because definition can have multiple keywords
export interface Rule {
  keyword: string
  definition: AddedKeywordDefinition
***REMOVED***

export function getRules(): ValidationRules {
  const groups: Record<"number" | "string" | "array" | "object", RuleGroup> = {
    number: {type: "number", rules: []***REMOVED***,
    string: {type: "string", rules: []***REMOVED***,
    array: {type: "array", rules: []***REMOVED***,
    object: {type: "object", rules: []***REMOVED***,
  ***REMOVED***
  return {
    types: {...groups, integer: true, boolean: true, null: true***REMOVED***,
    rules: [{rules: []***REMOVED***, groups.number, groups.string, groups.array, groups.object],
    post: {rules: []***REMOVED***,
    all: {***REMOVED***,
    keywords: {***REMOVED***,
  ***REMOVED***
***REMOVED***
