// on-finished/index.js
/*!
 * on-finished
 * Copyright(c) 2013 Jonathan Ong
 * Copyright(c) 2014 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module exports.
 * @public
 */

module.exports = onFinished
module.exports.isFinished = isFinished

/**
 * Module dependencies.
 * @private
 */

var asyncHooks = tryRequireAsyncHooks()
var first = require('ee-first')

/**
 * Variables.
 * @private
 */

/* istanbul ignore next */
var defer = typeof setImmediate === 'function'
  ? setImmediate
  : function (fn) { process.nextTick(fn.bind.apply(fn, arguments)) ***REMOVED***

/**
 * Invoke callback when the response has finished, useful for
 * cleaning up resources afterwards.
 *
 * @param {object***REMOVED*** msg
 * @param {function***REMOVED*** listener
 * @return {object***REMOVED***
 * @public
 */

function onFinished (msg, listener) {
  if (isFinished(msg) !== false) {
    defer(listener, null, msg)
    return msg
  ***REMOVED***

  // attach the listener to the message
  attachListener(msg, wrap(listener))

  return msg
***REMOVED***

/**
 * Determine if message is already finished.
 *
 * @param {object***REMOVED*** msg
 * @return {boolean***REMOVED***
 * @public
 */

function isFinished (msg) {
  var socket = msg.socket

  if (typeof msg.finished === 'boolean') {
    // OutgoingMessage
    return Boolean(msg.finished || (socket && !socket.writable))
  ***REMOVED***

  if (typeof msg.complete === 'boolean') {
    // IncomingMessage
    return Boolean(msg.upgrade || !socket || !socket.readable || (msg.complete && !msg.readable))
  ***REMOVED***

  // don't know
  return undefined
***REMOVED***

/**
 * Attach a finished listener to the message.
 *
 * @param {object***REMOVED*** msg
 * @param {function***REMOVED*** callback
 * @private
 */

function attachFinishedListener (msg, callback) {
  var eeMsg
  var eeSocket
  var finished = false

  function onFinish (error) {
    eeMsg.cancel()
    eeSocket.cancel()

    finished = true
    callback(error)
  ***REMOVED***

  // finished on first message event
  eeMsg = eeSocket = first([[msg, 'end', 'finish']], onFinish)

  function onSocket (socket) {
    // remove listener
    msg.removeListener('socket', onSocket)

    if (finished) return
    if (eeMsg !== eeSocket) return

    // finished on first socket event
    eeSocket = first([[socket, 'error', 'close']], onFinish)
  ***REMOVED***

  if (msg.socket) {
    // socket already assigned
    onSocket(msg.socket)
    return
  ***REMOVED***

  // wait for socket to be assigned
  msg.on('socket', onSocket)

  if (msg.socket === undefined) {
    // istanbul ignore next: node.js 0.8 patch
    patchAssignSocket(msg, onSocket)
  ***REMOVED***
***REMOVED***

/**
 * Attach the listener to the message.
 *
 * @param {object***REMOVED*** msg
 * @return {function***REMOVED***
 * @private
 */

function attachListener (msg, listener) {
  var attached = msg.__onFinished

  // create a private single listener with queue
  if (!attached || !attached.queue) {
    attached = msg.__onFinished = createListener(msg)
    attachFinishedListener(msg, attached)
  ***REMOVED***

  attached.queue.push(listener)
***REMOVED***

/**
 * Create listener on message.
 *
 * @param {object***REMOVED*** msg
 * @return {function***REMOVED***
 * @private
 */

function createListener (msg) {
  function listener (err) {
    if (msg.__onFinished === listener) msg.__onFinished = null
    if (!listener.queue) return

    var queue = listener.queue
    listener.queue = null

    for (var i = 0; i < queue.length; i++) {
      queue[i](err, msg)
    ***REMOVED***
  ***REMOVED***

  listener.queue = []

  return listener
***REMOVED***

/**
 * Patch ServerResponse.prototype.assignSocket for node.js 0.8.
 *
 * @param {ServerResponse***REMOVED*** res
 * @param {function***REMOVED*** callback
 * @private
 */

// istanbul ignore next: node.js 0.8 patch
function patchAssignSocket (res, callback) {
  var assignSocket = res.assignSocket

  if (typeof assignSocket !== 'function') return

  // res.on('socket', callback) is broken in 0.8
  res.assignSocket = function _assignSocket (socket) {
    assignSocket.call(this, socket)
    callback(socket)
  ***REMOVED***
***REMOVED***

/**
 * Try to require async_hooks
 * @private
 */

function tryRequireAsyncHooks () {
  try {
    return require('async_hooks')
  ***REMOVED*** catch (e) {
    return {***REMOVED***
  ***REMOVED***
***REMOVED***

/**
 * Wrap function with async resource, if possible.
 * AsyncResource.bind static method backported.
 * @private
 */

function wrap (fn) {
  var res

  // create anonymous resource
  if (asyncHooks.AsyncResource) {
    res = new asyncHooks.AsyncResource(fn.name || 'bound-anonymous-fn')
  ***REMOVED***

  // incompatible node.js
  if (!res || !res.runInAsyncScope) {
    return fn
  ***REMOVED***

  // return bound function
  return res.runInAsyncScope.bind(res, fn, null)
***REMOVED***
