// jest-config/build/Deprecated.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
***REMOVED***);
exports.default = void 0;

function _chalk() {
  const data = _interopRequireDefault(require('chalk'));

  _chalk = function () {
    return data;
  ***REMOVED***;

  return data;
***REMOVED***

function _prettyFormat() {
  const data = require('pretty-format');

  _prettyFormat = function () {
    return data;
  ***REMOVED***;

  return data;
***REMOVED***

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {default: obj***REMOVED***;
***REMOVED***

/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const format = value =>
  (0, _prettyFormat().format)(value, {
    min: true
  ***REMOVED***);

const deprecatedOptions = {
  browser: () =>
    `  Option ${_chalk().default.bold(
      '"browser"'
    )***REMOVED*** has been deprecated. Please install "browser-resolve" and use the "resolver" option in Jest configuration as shown in the documentation: https://jestjs.io/docs/configuration#resolver-string`,
  preprocessorIgnorePatterns: options => `  Option ${_chalk().default.bold(
    '"preprocessorIgnorePatterns"'
  )***REMOVED*** was replaced by ${_chalk().default.bold(
    '"transformIgnorePatterns"'
  )***REMOVED***, which support multiple preprocessors.

  Jest now treats your current configuration as:
  {
    ${_chalk().default.bold(
      '"transformIgnorePatterns"'
    )***REMOVED***: ${_chalk().default.bold(format(options.preprocessorIgnorePatterns))***REMOVED***
  ***REMOVED***

  Please update your configuration.`,
  scriptPreprocessor: options => `  Option ${_chalk().default.bold(
    '"scriptPreprocessor"'
  )***REMOVED*** was replaced by ${_chalk().default.bold(
    '"transform"'
  )***REMOVED***, which support multiple preprocessors.

  Jest now treats your current configuration as:
  {
    ${_chalk().default.bold('"transform"')***REMOVED***: ${_chalk().default.bold(
    `{".*": ${format(options.scriptPreprocessor)***REMOVED******REMOVED***`
  )***REMOVED***
  ***REMOVED***

  Please update your configuration.`,
  setupTestFrameworkScriptFile: _options => `  Option ${_chalk().default.bold(
    '"setupTestFrameworkScriptFile"'
  )***REMOVED*** was replaced by configuration ${_chalk().default.bold(
    '"setupFilesAfterEnv"'
  )***REMOVED***, which supports multiple paths.

  Please update your configuration.`,
  testPathDirs: options => `  Option ${_chalk().default.bold(
    '"testPathDirs"'
  )***REMOVED*** was replaced by ${_chalk().default.bold('"roots"')***REMOVED***.

  Jest now treats your current configuration as:
  {
    ${_chalk().default.bold('"roots"')***REMOVED***: ${_chalk().default.bold(
    format(options.testPathDirs)
  )***REMOVED***
  ***REMOVED***

  Please update your configuration.
  `
***REMOVED***;
var _default = deprecatedOptions;
exports.default = _default;
