// sucrase/dist/esm/parser/traverser/base.js
import State from "../tokenizer/state";
import {charCodes***REMOVED*** from "../util/charcodes";

export let isJSXEnabled;
export let isTypeScriptEnabled;
export let isFlowEnabled;
export let state;
export let input;
export let nextContextId;

export function getNextContextId() {
  return nextContextId++;
***REMOVED***

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function augmentError(error) {
  if ("pos" in error) {
    const loc = locationForIndex(error.pos);
    error.message += ` (${loc.line***REMOVED***:${loc.column***REMOVED***)`;
    error.loc = loc;
  ***REMOVED***
  return error;
***REMOVED***

export class Loc {
  
  
  constructor(line, column) {
    this.line = line;
    this.column = column;
  ***REMOVED***
***REMOVED***

export function locationForIndex(pos) {
  let line = 1;
  let column = 1;
  for (let i = 0; i < pos; i++) {
    if (input.charCodeAt(i) === charCodes.lineFeed) {
      line++;
      column = 1;
    ***REMOVED*** else {
      column++;
    ***REMOVED***
  ***REMOVED***
  return new Loc(line, column);
***REMOVED***

export function initParser(
  inputCode,
  isJSXEnabledArg,
  isTypeScriptEnabledArg,
  isFlowEnabledArg,
) {
  input = inputCode;
  state = new State();
  nextContextId = 1;
  isJSXEnabled = isJSXEnabledArg;
  isTypeScriptEnabled = isTypeScriptEnabledArg;
  isFlowEnabled = isFlowEnabledArg;
***REMOVED***
