// sucrase/dist/esm/parser/traverser/lval.js
import {flowParseAssignableListItemTypes***REMOVED*** from "../plugins/flow";
import {tsParseAssignableListItemTypes, tsParseModifiers***REMOVED*** from "../plugins/typescript";
import {
  eat,
  IdentifierRole,
  match,
  next,
  popTypeContext,
  pushTypeContext,
***REMOVED*** from "../tokenizer/index";
import {ContextualKeyword***REMOVED*** from "../tokenizer/keywords";
import {TokenType, TokenType as tt***REMOVED*** from "../tokenizer/types";
import {isFlowEnabled, isTypeScriptEnabled, state***REMOVED*** from "./base";
import {parseIdentifier, parseMaybeAssign, parseObj***REMOVED*** from "./expression";
import {expect, unexpected***REMOVED*** from "./util";

export function parseSpread() {
  next();
  parseMaybeAssign(false);
***REMOVED***

export function parseRest(isBlockScope) {
  next();
  parseBindingAtom(isBlockScope);
***REMOVED***

export function parseBindingIdentifier(isBlockScope) {
  parseIdentifier();
  markPriorBindingIdentifier(isBlockScope);
***REMOVED***

export function parseImportedIdentifier() {
  parseIdentifier();
  state.tokens[state.tokens.length - 1].identifierRole = IdentifierRole.ImportDeclaration;
***REMOVED***

export function markPriorBindingIdentifier(isBlockScope) {
  let identifierRole;
  if (state.scopeDepth === 0) {
    identifierRole = IdentifierRole.TopLevelDeclaration;
  ***REMOVED*** else if (isBlockScope) {
    identifierRole = IdentifierRole.BlockScopedDeclaration;
  ***REMOVED*** else {
    identifierRole = IdentifierRole.FunctionScopedDeclaration;
  ***REMOVED***
  state.tokens[state.tokens.length - 1].identifierRole = identifierRole;
***REMOVED***

// Parses lvalue (assignable) atom.
export function parseBindingAtom(isBlockScope) {
  switch (state.type) {
    case tt._this: {
      // In TypeScript, "this" may be the name of a parameter, so allow it.
      const oldIsType = pushTypeContext(0);
      next();
      popTypeContext(oldIsType);
      return;
    ***REMOVED***

    case tt._yield:
    case tt.name: {
      state.type = tt.name;
      parseBindingIdentifier(isBlockScope);
      return;
    ***REMOVED***

    case tt.bracketL: {
      next();
      parseBindingList(tt.bracketR, isBlockScope, true /* allowEmpty */);
      return;
    ***REMOVED***

    case tt.braceL:
      parseObj(true, isBlockScope);
      return;

    default:
      unexpected();
  ***REMOVED***
***REMOVED***

export function parseBindingList(
  close,
  isBlockScope,
  allowEmpty = false,
  allowModifiers = false,
  contextId = 0,
) {
  let first = true;

  let hasRemovedComma = false;
  const firstItemTokenIndex = state.tokens.length;

  while (!eat(close) && !state.error) {
    if (first) {
      first = false;
    ***REMOVED*** else {
      expect(tt.comma);
      state.tokens[state.tokens.length - 1].contextId = contextId;
      // After a "this" type in TypeScript, we need to set the following comma (if any) to also be
      // a type token so that it will be removed.
      if (!hasRemovedComma && state.tokens[firstItemTokenIndex].isType) {
        state.tokens[state.tokens.length - 1].isType = true;
        hasRemovedComma = true;
      ***REMOVED***
    ***REMOVED***
    if (allowEmpty && match(tt.comma)) {
      // Empty item; nothing further to parse for this item.
    ***REMOVED*** else if (eat(close)) {
      break;
    ***REMOVED*** else if (match(tt.ellipsis)) {
      parseRest(isBlockScope);
      parseAssignableListItemTypes();
      // Support rest element trailing commas allowed by TypeScript <2.9.
      eat(TokenType.comma);
      expect(close);
      break;
    ***REMOVED*** else {
      parseAssignableListItem(allowModifiers, isBlockScope);
    ***REMOVED***
  ***REMOVED***
***REMOVED***

function parseAssignableListItem(allowModifiers, isBlockScope) {
  if (allowModifiers) {
    tsParseModifiers([
      ContextualKeyword._public,
      ContextualKeyword._protected,
      ContextualKeyword._private,
      ContextualKeyword._readonly,
      ContextualKeyword._override,
    ]);
  ***REMOVED***

  parseMaybeDefault(isBlockScope);
  parseAssignableListItemTypes();
  parseMaybeDefault(isBlockScope, true /* leftAlreadyParsed */);
***REMOVED***

function parseAssignableListItemTypes() {
  if (isFlowEnabled) {
    flowParseAssignableListItemTypes();
  ***REMOVED*** else if (isTypeScriptEnabled) {
    tsParseAssignableListItemTypes();
  ***REMOVED***
***REMOVED***

// Parses assignment pattern around given atom if possible.
export function parseMaybeDefault(isBlockScope, leftAlreadyParsed = false) {
  if (!leftAlreadyParsed) {
    parseBindingAtom(isBlockScope);
  ***REMOVED***
  if (!eat(tt.eq)) {
    return;
  ***REMOVED***
  const eqIndex = state.tokens.length - 1;
  parseMaybeAssign();
  state.tokens[eqIndex].rhsEndIndex = state.tokens.length;
***REMOVED***
