// sucrase/dist/esm/parser/traverser/statement.js
/* eslint max-len: 0 */

import {File***REMOVED*** from "../index";
import {
  flowAfterParseClassSuper,
  flowAfterParseVarHead,
  flowParseExportDeclaration,
  flowParseExportStar,
  flowParseIdentifierStatement,
  flowParseImportSpecifier,
  flowParseTypeAnnotation,
  flowParseTypeParameterDeclaration,
  flowShouldDisallowExportDefaultSpecifier,
  flowShouldParseExportDeclaration,
  flowShouldParseExportStar,
  flowStartParseFunctionParams,
  flowStartParseImportSpecifiers,
  flowTryParseExportDefaultExpression,
  flowTryParseStatement,
***REMOVED*** from "../plugins/flow";
import {
  tsAfterParseClassSuper,
  tsAfterParseVarHead,
  tsIsDeclarationStart,
  tsParseExportDeclaration,
  tsParseExportSpecifier,
  tsParseIdentifierStatement,
  tsParseImportEqualsDeclaration,
  tsParseImportSpecifier,
  tsParseMaybeDecoratorArguments,
  tsParseModifiers,
  tsStartParseFunctionParams,
  tsTryParseClassMemberWithIsStatic,
  tsTryParseExport,
  tsTryParseExportDefaultExpression,
  tsTryParseStatementContent,
  tsTryParseTypeAnnotation,
  tsTryParseTypeParameters,
***REMOVED*** from "../plugins/typescript";
import {
  eat,
  eatTypeToken,
  IdentifierRole,
  lookaheadType,
  lookaheadTypeAndKeyword,
  match,
  next,
  nextTokenStart,
  nextTokenStartSince,
  popTypeContext,
  pushTypeContext,
***REMOVED*** from "../tokenizer";
import {ContextualKeyword***REMOVED*** from "../tokenizer/keywords";
import {Scope***REMOVED*** from "../tokenizer/state";
import { TokenType as tt***REMOVED*** from "../tokenizer/types";
import {charCodes***REMOVED*** from "../util/charcodes";
import {getNextContextId, input, isFlowEnabled, isTypeScriptEnabled, state***REMOVED*** from "./base";
import {
  parseCallExpressionArguments,
  parseExprAtom,
  parseExpression,
  parseExprSubscripts,
  parseFunctionBodyAndFinish,
  parseIdentifier,
  parseMaybeAssign,
  parseMethod,
  parseObj,
  parseParenExpression,
  parsePropertyName,
***REMOVED*** from "./expression";
import {
  parseBindingAtom,
  parseBindingIdentifier,
  parseBindingList,
  parseImportedIdentifier,
***REMOVED*** from "./lval";
import {
  canInsertSemicolon,
  eatContextual,
  expect,
  expectContextual,
  hasFollowingLineBreak,
  hasPrecedingLineBreak,
  isContextual,
  isLineTerminator,
  isLookaheadContextual,
  semicolon,
  unexpected,
***REMOVED*** from "./util";

export function parseTopLevel() {
  parseBlockBody(tt.eof);
  state.scopes.push(new Scope(0, state.tokens.length, true));
  if (state.scopeDepth !== 0) {
    throw new Error(`Invalid scope depth at end of file: ${state.scopeDepth***REMOVED***`);
  ***REMOVED***
  return new File(state.tokens, state.scopes);
***REMOVED***

// Parse a single statement.
//
// If expecting a statement and finding a slash operator, parse a
// regular expression literal. This is to handle cases like
// `if (foo) /blah/.exec(foo)`, where looking at the previous token
// does not help.

export function parseStatement(declaration) {
  if (isFlowEnabled) {
    if (flowTryParseStatement()) {
      return;
    ***REMOVED***
  ***REMOVED***
  if (match(tt.at)) {
    parseDecorators();
  ***REMOVED***
  parseStatementContent(declaration);
***REMOVED***

function parseStatementContent(declaration) {
  if (isTypeScriptEnabled) {
    if (tsTryParseStatementContent()) {
      return;
    ***REMOVED***
  ***REMOVED***

  const starttype = state.type;

  // Most types of statements are recognized by the keyword they
  // start with. Many are trivial to parse, some require a bit of
  // complexity.

  switch (starttype) {
    case tt._break:
    case tt._continue:
      parseBreakContinueStatement();
      return;
    case tt._debugger:
      parseDebuggerStatement();
      return;
    case tt._do:
      parseDoStatement();
      return;
    case tt._for:
      parseForStatement();
      return;
    case tt._function:
      if (lookaheadType() === tt.dot) break;
      if (!declaration) unexpected();
      parseFunctionStatement();
      return;

    case tt._class:
      if (!declaration) unexpected();
      parseClass(true);
      return;

    case tt._if:
      parseIfStatement();
      return;
    case tt._return:
      parseReturnStatement();
      return;
    case tt._switch:
      parseSwitchStatement();
      return;
    case tt._throw:
      parseThrowStatement();
      return;
    case tt._try:
      parseTryStatement();
      return;

    case tt._let:
    case tt._const:
      if (!declaration) unexpected(); // NOTE: falls through to _var

    case tt._var:
      parseVarStatement(starttype !== tt._var);
      return;

    case tt._while:
      parseWhileStatement();
      return;
    case tt.braceL:
      parseBlock();
      return;
    case tt.semi:
      parseEmptyStatement();
      return;
    case tt._export:
    case tt._import: {
      const nextType = lookaheadType();
      if (nextType === tt.parenL || nextType === tt.dot) {
        break;
      ***REMOVED***
      next();
      if (starttype === tt._import) {
        parseImport();
      ***REMOVED*** else {
        parseExport();
      ***REMOVED***
      return;
    ***REMOVED***
    case tt.name:
      if (state.contextualKeyword === ContextualKeyword._async) {
        const functionStart = state.start;
        // peek ahead and see if next token is a function
        const snapshot = state.snapshot();
        next();
        if (match(tt._function) && !canInsertSemicolon()) {
          expect(tt._function);
          parseFunction(functionStart, true);
          return;
        ***REMOVED*** else {
          state.restoreFromSnapshot(snapshot);
        ***REMOVED***
      ***REMOVED*** else if (
        state.contextualKeyword === ContextualKeyword._using &&
        !hasFollowingLineBreak() &&
        // Statements like `using[0]` and `using in foo` aren't actual using
        // declarations.
        lookaheadType() === tt.name
      ) {
        parseVarStatement(true);
        return;
      ***REMOVED*** else if (startsAwaitUsing()) {
        expectContextual(ContextualKeyword._await);
        parseVarStatement(true);
        return;
      ***REMOVED***
    default:
      // Do nothing.
      break;
  ***REMOVED***

  // If the statement does not start with a statement keyword or a
  // brace, it's an ExpressionStatement or LabeledStatement. We
  // simply start parsing an expression, and afterwards, if the
  // next token is a colon and the expression was a simple
  // Identifier node, we switch to interpreting it as a label.
  const initialTokensLength = state.tokens.length;
  parseExpression();
  let simpleName = null;
  if (state.tokens.length === initialTokensLength + 1) {
    const token = state.tokens[state.tokens.length - 1];
    if (token.type === tt.name) {
      simpleName = token.contextualKeyword;
    ***REMOVED***
  ***REMOVED***
  if (simpleName == null) {
    semicolon();
    return;
  ***REMOVED***
  if (eat(tt.colon)) {
    parseLabeledStatement();
  ***REMOVED*** else {
    // This was an identifier, so we might want to handle flow/typescript-specific cases.
    parseIdentifierStatement(simpleName);
  ***REMOVED***
***REMOVED***

/**
 * Determine if we're positioned at an `await using` declaration.
 *
 * Note that this can happen either in place of a regular variable declaration
 * or in a loop body, and in both places, there are similar-looking cases where
 * we need to return false.
 *
 * Examples returning true:
 * await using foo = bar();
 * for (await using a of b) {***REMOVED***
 *
 * Examples returning false:
 * await using
 * await using + 1
 * await using instanceof T
 * for (await using;;) {***REMOVED***
 *
 * For now, we early return if we don't see `await`, then do a simple
 * backtracking-based lookahead for the `using` and identifier tokens. In the
 * future, this could be optimized with a character-based approach.
 */
function startsAwaitUsing() {
  if (!isContextual(ContextualKeyword._await)) {
    return false;
  ***REMOVED***
  const snapshot = state.snapshot();
  // await
  next();
  if (!isContextual(ContextualKeyword._using) || hasPrecedingLineBreak()) {
    state.restoreFromSnapshot(snapshot);
    return false;
  ***REMOVED***
  // using
  next();
  if (!match(tt.name) || hasPrecedingLineBreak()) {
    state.restoreFromSnapshot(snapshot);
    return false;
  ***REMOVED***
  state.restoreFromSnapshot(snapshot);
  return true;
***REMOVED***

export function parseDecorators() {
  while (match(tt.at)) {
    parseDecorator();
  ***REMOVED***
***REMOVED***

function parseDecorator() {
  next();
  if (eat(tt.parenL)) {
    parseExpression();
    expect(tt.parenR);
  ***REMOVED*** else {
    parseIdentifier();
    while (eat(tt.dot)) {
      parseIdentifier();
    ***REMOVED***
    parseMaybeDecoratorArguments();
  ***REMOVED***
***REMOVED***

function parseMaybeDecoratorArguments() {
  if (isTypeScriptEnabled) {
    tsParseMaybeDecoratorArguments();
  ***REMOVED*** else {
    baseParseMaybeDecoratorArguments();
  ***REMOVED***
***REMOVED***

export function baseParseMaybeDecoratorArguments() {
  if (eat(tt.parenL)) {
    parseCallExpressionArguments();
  ***REMOVED***
***REMOVED***

function parseBreakContinueStatement() {
  next();
  if (!isLineTerminator()) {
    parseIdentifier();
    semicolon();
  ***REMOVED***
***REMOVED***

function parseDebuggerStatement() {
  next();
  semicolon();
***REMOVED***

function parseDoStatement() {
  next();
  parseStatement(false);
  expect(tt._while);
  parseParenExpression();
  eat(tt.semi);
***REMOVED***

function parseForStatement() {
  state.scopeDepth++;
  const startTokenIndex = state.tokens.length;
  parseAmbiguousForStatement();
  const endTokenIndex = state.tokens.length;
  state.scopes.push(new Scope(startTokenIndex, endTokenIndex, false));
  state.scopeDepth--;
***REMOVED***

/**
 * Determine if this token is a `using` declaration (explicit resource
 * management) as part of a loop.
 * https://github.com/tc39/proposal-explicit-resource-management
 */
function isUsingInLoop() {
  if (!isContextual(ContextualKeyword._using)) {
    return false;
  ***REMOVED***
  // This must be `for (using of`, where `using` is the name of the loop
  // variable.
  if (isLookaheadContextual(ContextualKeyword._of)) {
    return false;
  ***REMOVED***
  return true;
***REMOVED***

// Disambiguating between a `for` and a `for`/`in` or `for`/`of`
// loop is non-trivial. Basically, we have to parse the init `var`
// statement or expression, disallowing the `in` operator (see
// the second parameter to `parseExpression`), and then check
// whether the next token is `in` or `of`. When there is no init
// part (semicolon immediately after the opening parenthesis), it
// is a regular `for` loop.
function parseAmbiguousForStatement() {
  next();

  let forAwait = false;
  if (isContextual(ContextualKeyword._await)) {
    forAwait = true;
    next();
  ***REMOVED***
  expect(tt.parenL);

  if (match(tt.semi)) {
    if (forAwait) {
      unexpected();
    ***REMOVED***
    parseFor();
    return;
  ***REMOVED***

  const isAwaitUsing = startsAwaitUsing();
  if (isAwaitUsing || match(tt._var) || match(tt._let) || match(tt._const) || isUsingInLoop()) {
    if (isAwaitUsing) {
      expectContextual(ContextualKeyword._await);
    ***REMOVED***
    next();
    parseVar(true, state.type !== tt._var);
    if (match(tt._in) || isContextual(ContextualKeyword._of)) {
      parseForIn(forAwait);
      return;
    ***REMOVED***
    parseFor();
    return;
  ***REMOVED***

  parseExpression(true);
  if (match(tt._in) || isContextual(ContextualKeyword._of)) {
    parseForIn(forAwait);
    return;
  ***REMOVED***
  if (forAwait) {
    unexpected();
  ***REMOVED***
  parseFor();
***REMOVED***

function parseFunctionStatement() {
  const functionStart = state.start;
  next();
  parseFunction(functionStart, true);
***REMOVED***

function parseIfStatement() {
  next();
  parseParenExpression();
  parseStatement(false);
  if (eat(tt._else)) {
    parseStatement(false);
  ***REMOVED***
***REMOVED***

function parseReturnStatement() {
  next();

  // In `return` (and `break`/`continue`), the keywords with
  // optional arguments, we eagerly look for a semicolon or the
  // possibility to insert one.

  if (!isLineTerminator()) {
    parseExpression();
    semicolon();
  ***REMOVED***
***REMOVED***

function parseSwitchStatement() {
  next();
  parseParenExpression();
  state.scopeDepth++;
  const startTokenIndex = state.tokens.length;
  expect(tt.braceL);

  // Don't bother validation; just go through any sequence of cases, defaults, and statements.
  while (!match(tt.braceR) && !state.error) {
    if (match(tt._case) || match(tt._default)) {
      const isCase = match(tt._case);
      next();
      if (isCase) {
        parseExpression();
      ***REMOVED***
      expect(tt.colon);
    ***REMOVED*** else {
      parseStatement(true);
    ***REMOVED***
  ***REMOVED***
  next(); // Closing brace
  const endTokenIndex = state.tokens.length;
  state.scopes.push(new Scope(startTokenIndex, endTokenIndex, false));
  state.scopeDepth--;
***REMOVED***

function parseThrowStatement() {
  next();
  parseExpression();
  semicolon();
***REMOVED***

function parseCatchClauseParam() {
  parseBindingAtom(true /* isBlockScope */);

  if (isTypeScriptEnabled) {
    tsTryParseTypeAnnotation();
  ***REMOVED***
***REMOVED***

function parseTryStatement() {
  next();

  parseBlock();

  if (match(tt._catch)) {
    next();
    let catchBindingStartTokenIndex = null;
    if (match(tt.parenL)) {
      state.scopeDepth++;
      catchBindingStartTokenIndex = state.tokens.length;
      expect(tt.parenL);
      parseCatchClauseParam();
      expect(tt.parenR);
    ***REMOVED***
    parseBlock();
    if (catchBindingStartTokenIndex != null) {
      // We need a special scope for the catch binding which includes the binding itself and the
      // catch block.
      const endTokenIndex = state.tokens.length;
      state.scopes.push(new Scope(catchBindingStartTokenIndex, endTokenIndex, false));
      state.scopeDepth--;
    ***REMOVED***
  ***REMOVED***
  if (eat(tt._finally)) {
    parseBlock();
  ***REMOVED***
***REMOVED***

export function parseVarStatement(isBlockScope) {
  next();
  parseVar(false, isBlockScope);
  semicolon();
***REMOVED***

function parseWhileStatement() {
  next();
  parseParenExpression();
  parseStatement(false);
***REMOVED***

function parseEmptyStatement() {
  next();
***REMOVED***

function parseLabeledStatement() {
  parseStatement(true);
***REMOVED***

/**
 * Parse a statement starting with an identifier of the given name. Subclasses match on the name
 * to handle statements like "declare".
 */
function parseIdentifierStatement(contextualKeyword) {
  if (isTypeScriptEnabled) {
    tsParseIdentifierStatement(contextualKeyword);
  ***REMOVED*** else if (isFlowEnabled) {
    flowParseIdentifierStatement(contextualKeyword);
  ***REMOVED*** else {
    semicolon();
  ***REMOVED***
***REMOVED***

// Parse a semicolon-enclosed block of statements.
export function parseBlock(isFunctionScope = false, contextId = 0) {
  const startTokenIndex = state.tokens.length;
  state.scopeDepth++;
  expect(tt.braceL);
  if (contextId) {
    state.tokens[state.tokens.length - 1].contextId = contextId;
  ***REMOVED***
  parseBlockBody(tt.braceR);
  if (contextId) {
    state.tokens[state.tokens.length - 1].contextId = contextId;
  ***REMOVED***
  const endTokenIndex = state.tokens.length;
  state.scopes.push(new Scope(startTokenIndex, endTokenIndex, isFunctionScope));
  state.scopeDepth--;
***REMOVED***

export function parseBlockBody(end) {
  while (!eat(end) && !state.error) {
    parseStatement(true);
  ***REMOVED***
***REMOVED***

// Parse a regular `for` loop. The disambiguation code in
// `parseStatement` will already have parsed the init statement or
// expression.

function parseFor() {
  expect(tt.semi);
  if (!match(tt.semi)) {
    parseExpression();
  ***REMOVED***
  expect(tt.semi);
  if (!match(tt.parenR)) {
    parseExpression();
  ***REMOVED***
  expect(tt.parenR);
  parseStatement(false);
***REMOVED***

// Parse a `for`/`in` and `for`/`of` loop, which are almost
// same from parser's perspective.

function parseForIn(forAwait) {
  if (forAwait) {
    eatContextual(ContextualKeyword._of);
  ***REMOVED*** else {
    next();
  ***REMOVED***
  parseExpression();
  expect(tt.parenR);
  parseStatement(false);
***REMOVED***

// Parse a list of variable declarations.

function parseVar(isFor, isBlockScope) {
  while (true) {
    parseVarHead(isBlockScope);
    if (eat(tt.eq)) {
      const eqIndex = state.tokens.length - 1;
      parseMaybeAssign(isFor);
      state.tokens[eqIndex].rhsEndIndex = state.tokens.length;
    ***REMOVED***
    if (!eat(tt.comma)) {
      break;
    ***REMOVED***
  ***REMOVED***
***REMOVED***

function parseVarHead(isBlockScope) {
  parseBindingAtom(isBlockScope);
  if (isTypeScriptEnabled) {
    tsAfterParseVarHead();
  ***REMOVED*** else if (isFlowEnabled) {
    flowAfterParseVarHead();
  ***REMOVED***
***REMOVED***

// Parse a function declaration or literal (depending on the
// `isStatement` parameter).

export function parseFunction(
  functionStart,
  isStatement,
  optionalId = false,
) {
  if (match(tt.star)) {
    next();
  ***REMOVED***

  if (isStatement && !optionalId && !match(tt.name) && !match(tt._yield)) {
    unexpected();
  ***REMOVED***

  let nameScopeStartTokenIndex = null;

  if (match(tt.name)) {
    // Expression-style functions should limit their name's scope to the function body, so we make
    // a new function scope to enforce that.
    if (!isStatement) {
      nameScopeStartTokenIndex = state.tokens.length;
      state.scopeDepth++;
    ***REMOVED***
    parseBindingIdentifier(false);
  ***REMOVED***

  const startTokenIndex = state.tokens.length;
  state.scopeDepth++;
  parseFunctionParams();
  parseFunctionBodyAndFinish(functionStart);
  const endTokenIndex = state.tokens.length;
  // In addition to the block scope of the function body, we need a separate function-style scope
  // that includes the params.
  state.scopes.push(new Scope(startTokenIndex, endTokenIndex, true));
  state.scopeDepth--;
  if (nameScopeStartTokenIndex !== null) {
    state.scopes.push(new Scope(nameScopeStartTokenIndex, endTokenIndex, true));
    state.scopeDepth--;
  ***REMOVED***
***REMOVED***

export function parseFunctionParams(
  allowModifiers = false,
  funcContextId = 0,
) {
  if (isTypeScriptEnabled) {
    tsStartParseFunctionParams();
  ***REMOVED*** else if (isFlowEnabled) {
    flowStartParseFunctionParams();
  ***REMOVED***

  expect(tt.parenL);
  if (funcContextId) {
    state.tokens[state.tokens.length - 1].contextId = funcContextId;
  ***REMOVED***
  parseBindingList(
    tt.parenR,
    false /* isBlockScope */,
    false /* allowEmpty */,
    allowModifiers,
    funcContextId,
  );
  if (funcContextId) {
    state.tokens[state.tokens.length - 1].contextId = funcContextId;
  ***REMOVED***
***REMOVED***

// Parse a class declaration or literal (depending on the
// `isStatement` parameter).

export function parseClass(isStatement, optionalId = false) {
  // Put a context ID on the class keyword, the open-brace, and the close-brace, so that later
  // code can easily navigate to meaningful points on the class.
  const contextId = getNextContextId();

  next();
  state.tokens[state.tokens.length - 1].contextId = contextId;
  state.tokens[state.tokens.length - 1].isExpression = !isStatement;
  // Like with functions, we declare a special "name scope" from the start of the name to the end
  // of the class, but only with expression-style classes, to represent the fact that the name is
  // available to the body of the class but not an outer declaration.
  let nameScopeStartTokenIndex = null;
  if (!isStatement) {
    nameScopeStartTokenIndex = state.tokens.length;
    state.scopeDepth++;
  ***REMOVED***
  parseClassId(isStatement, optionalId);
  parseClassSuper();
  const openBraceIndex = state.tokens.length;
  parseClassBody(contextId);
  if (state.error) {
    return;
  ***REMOVED***
  state.tokens[openBraceIndex].contextId = contextId;
  state.tokens[state.tokens.length - 1].contextId = contextId;
  if (nameScopeStartTokenIndex !== null) {
    const endTokenIndex = state.tokens.length;
    state.scopes.push(new Scope(nameScopeStartTokenIndex, endTokenIndex, false));
    state.scopeDepth--;
  ***REMOVED***
***REMOVED***

function isClassProperty() {
  return match(tt.eq) || match(tt.semi) || match(tt.braceR) || match(tt.bang) || match(tt.colon);
***REMOVED***

function isClassMethod() {
  return match(tt.parenL) || match(tt.lessThan);
***REMOVED***

function parseClassBody(classContextId) {
  expect(tt.braceL);

  while (!eat(tt.braceR) && !state.error) {
    if (eat(tt.semi)) {
      continue;
    ***REMOVED***

    if (match(tt.at)) {
      parseDecorator();
      continue;
    ***REMOVED***
    const memberStart = state.start;
    parseClassMember(memberStart, classContextId);
  ***REMOVED***
***REMOVED***

function parseClassMember(memberStart, classContextId) {
  if (isTypeScriptEnabled) {
    tsParseModifiers([
      ContextualKeyword._declare,
      ContextualKeyword._public,
      ContextualKeyword._protected,
      ContextualKeyword._private,
      ContextualKeyword._override,
    ]);
  ***REMOVED***
  let isStatic = false;
  if (match(tt.name) && state.contextualKeyword === ContextualKeyword._static) {
    parseIdentifier(); // eats 'static'
    if (isClassMethod()) {
      parseClassMethod(memberStart, /* isConstructor */ false);
      return;
    ***REMOVED*** else if (isClassProperty()) {
      parseClassProperty();
      return;
    ***REMOVED***
    // otherwise something static
    state.tokens[state.tokens.length - 1].type = tt._static;
    isStatic = true;

    if (match(tt.braceL)) {
      // This is a static block. Mark the word "static" with the class context ID for class element
      // detection and parse as a regular block.
      state.tokens[state.tokens.length - 1].contextId = classContextId;
      parseBlock();
      return;
    ***REMOVED***
  ***REMOVED***

  parseClassMemberWithIsStatic(memberStart, isStatic, classContextId);
***REMOVED***

function parseClassMemberWithIsStatic(
  memberStart,
  isStatic,
  classContextId,
) {
  if (isTypeScriptEnabled) {
    if (tsTryParseClassMemberWithIsStatic(isStatic)) {
      return;
    ***REMOVED***
  ***REMOVED***
  if (eat(tt.star)) {
    // a generator
    parseClassPropertyName(classContextId);
    parseClassMethod(memberStart, /* isConstructor */ false);
    return;
  ***REMOVED***

  // Get the identifier name so we can tell if it's actually a keyword like "async", "get", or
  // "set".
  parseClassPropertyName(classContextId);
  let isConstructor = false;
  const token = state.tokens[state.tokens.length - 1];
  // We allow "constructor" as either an identifier or a string.
  if (token.contextualKeyword === ContextualKeyword._constructor) {
    isConstructor = true;
  ***REMOVED***
  parsePostMemberNameModifiers();

  if (isClassMethod()) {
    parseClassMethod(memberStart, isConstructor);
  ***REMOVED*** else if (isClassProperty()) {
    parseClassProperty();
  ***REMOVED*** else if (token.contextualKeyword === ContextualKeyword._async && !isLineTerminator()) {
    state.tokens[state.tokens.length - 1].type = tt._async;
    // an async method
    const isGenerator = match(tt.star);
    if (isGenerator) {
      next();
    ***REMOVED***

    // The so-called parsed name would have been "async": get the real name.
    parseClassPropertyName(classContextId);
    parsePostMemberNameModifiers();
    parseClassMethod(memberStart, false /* isConstructor */);
  ***REMOVED*** else if (
    (token.contextualKeyword === ContextualKeyword._get ||
      token.contextualKeyword === ContextualKeyword._set) &&
    !(isLineTerminator() && match(tt.star))
  ) {
    if (token.contextualKeyword === ContextualKeyword._get) {
      state.tokens[state.tokens.length - 1].type = tt._get;
    ***REMOVED*** else {
      state.tokens[state.tokens.length - 1].type = tt._set;
    ***REMOVED***
    // `get\n*` is an uninitialized property named 'get' followed by a generator.
    // a getter or setter
    // The so-called parsed name would have been "get/set": get the real name.
    parseClassPropertyName(classContextId);
    parseClassMethod(memberStart, /* isConstructor */ false);
  ***REMOVED*** else if (token.contextualKeyword === ContextualKeyword._accessor && !isLineTerminator()) {
    parseClassPropertyName(classContextId);
    parseClassProperty();
  ***REMOVED*** else if (isLineTerminator()) {
    // an uninitialized class property (due to ASI, since we don't otherwise recognize the next token)
    parseClassProperty();
  ***REMOVED*** else {
    unexpected();
  ***REMOVED***
***REMOVED***

function parseClassMethod(functionStart, isConstructor) {
  if (isTypeScriptEnabled) {
    tsTryParseTypeParameters();
  ***REMOVED*** else if (isFlowEnabled) {
    if (match(tt.lessThan)) {
      flowParseTypeParameterDeclaration();
    ***REMOVED***
  ***REMOVED***
  parseMethod(functionStart, isConstructor);
***REMOVED***

// Return the name of the class property, if it is a simple identifier.
export function parseClassPropertyName(classContextId) {
  parsePropertyName(classContextId);
***REMOVED***

export function parsePostMemberNameModifiers() {
  if (isTypeScriptEnabled) {
    const oldIsType = pushTypeContext(0);
    eat(tt.question);
    popTypeContext(oldIsType);
  ***REMOVED***
***REMOVED***

export function parseClassProperty() {
  if (isTypeScriptEnabled) {
    eatTypeToken(tt.bang);
    tsTryParseTypeAnnotation();
  ***REMOVED*** else if (isFlowEnabled) {
    if (match(tt.colon)) {
      flowParseTypeAnnotation();
    ***REMOVED***
  ***REMOVED***

  if (match(tt.eq)) {
    const equalsTokenIndex = state.tokens.length;
    next();
    parseMaybeAssign();
    state.tokens[equalsTokenIndex].rhsEndIndex = state.tokens.length;
  ***REMOVED***
  semicolon();
***REMOVED***

function parseClassId(isStatement, optionalId = false) {
  if (
    isTypeScriptEnabled &&
    (!isStatement || optionalId) &&
    isContextual(ContextualKeyword._implements)
  ) {
    return;
  ***REMOVED***

  if (match(tt.name)) {
    parseBindingIdentifier(true);
  ***REMOVED***

  if (isTypeScriptEnabled) {
    tsTryParseTypeParameters();
  ***REMOVED*** else if (isFlowEnabled) {
    if (match(tt.lessThan)) {
      flowParseTypeParameterDeclaration();
    ***REMOVED***
  ***REMOVED***
***REMOVED***

// Returns true if there was a superclass.
function parseClassSuper() {
  let hasSuper = false;
  if (eat(tt._extends)) {
    parseExprSubscripts();
    hasSuper = true;
  ***REMOVED*** else {
    hasSuper = false;
  ***REMOVED***
  if (isTypeScriptEnabled) {
    tsAfterParseClassSuper(hasSuper);
  ***REMOVED*** else if (isFlowEnabled) {
    flowAfterParseClassSuper(hasSuper);
  ***REMOVED***
***REMOVED***

// Parses module export declaration.

export function parseExport() {
  const exportIndex = state.tokens.length - 1;
  if (isTypeScriptEnabled) {
    if (tsTryParseExport()) {
      return;
    ***REMOVED***
  ***REMOVED***
  // export * from '...'
  if (shouldParseExportStar()) {
    parseExportStar();
  ***REMOVED*** else if (isExportDefaultSpecifier()) {
    // export default from
    parseIdentifier();
    if (match(tt.comma) && lookaheadType() === tt.star) {
      expect(tt.comma);
      expect(tt.star);
      expectContextual(ContextualKeyword._as);
      parseIdentifier();
    ***REMOVED*** else {
      parseExportSpecifiersMaybe();
    ***REMOVED***
    parseExportFrom();
  ***REMOVED*** else if (eat(tt._default)) {
    // export default ...
    parseExportDefaultExpression();
  ***REMOVED*** else if (shouldParseExportDeclaration()) {
    parseExportDeclaration();
  ***REMOVED*** else {
    // export { x, y as z ***REMOVED*** [from '...']
    parseExportSpecifiers();
    parseExportFrom();
  ***REMOVED***
  state.tokens[exportIndex].rhsEndIndex = state.tokens.length;
***REMOVED***

function parseExportDefaultExpression() {
  if (isTypeScriptEnabled) {
    if (tsTryParseExportDefaultExpression()) {
      return;
    ***REMOVED***
  ***REMOVED***
  if (isFlowEnabled) {
    if (flowTryParseExportDefaultExpression()) {
      return;
    ***REMOVED***
  ***REMOVED***
  const functionStart = state.start;
  if (eat(tt._function)) {
    parseFunction(functionStart, true, true);
  ***REMOVED*** else if (isContextual(ContextualKeyword._async) && lookaheadType() === tt._function) {
    // async function declaration
    eatContextual(ContextualKeyword._async);
    eat(tt._function);
    parseFunction(functionStart, true, true);
  ***REMOVED*** else if (match(tt._class)) {
    parseClass(true, true);
  ***REMOVED*** else if (match(tt.at)) {
    parseDecorators();
    parseClass(true, true);
  ***REMOVED*** else {
    parseMaybeAssign();
    semicolon();
  ***REMOVED***
***REMOVED***

function parseExportDeclaration() {
  if (isTypeScriptEnabled) {
    tsParseExportDeclaration();
  ***REMOVED*** else if (isFlowEnabled) {
    flowParseExportDeclaration();
  ***REMOVED*** else {
    parseStatement(true);
  ***REMOVED***
***REMOVED***

function isExportDefaultSpecifier() {
  if (isTypeScriptEnabled && tsIsDeclarationStart()) {
    return false;
  ***REMOVED*** else if (isFlowEnabled && flowShouldDisallowExportDefaultSpecifier()) {
    return false;
  ***REMOVED***
  if (match(tt.name)) {
    return state.contextualKeyword !== ContextualKeyword._async;
  ***REMOVED***

  if (!match(tt._default)) {
    return false;
  ***REMOVED***

  const _next = nextTokenStart();
  const lookahead = lookaheadTypeAndKeyword();
  const hasFrom =
    lookahead.type === tt.name && lookahead.contextualKeyword === ContextualKeyword._from;
  if (lookahead.type === tt.comma) {
    return true;
  ***REMOVED***
  // lookahead again when `export default from` is seen
  if (hasFrom) {
    const nextAfterFrom = input.charCodeAt(nextTokenStartSince(_next + 4));
    return nextAfterFrom === charCodes.quotationMark || nextAfterFrom === charCodes.apostrophe;
  ***REMOVED***
  return false;
***REMOVED***

function parseExportSpecifiersMaybe() {
  if (eat(tt.comma)) {
    parseExportSpecifiers();
  ***REMOVED***
***REMOVED***

export function parseExportFrom() {
  if (eatContextual(ContextualKeyword._from)) {
    parseExprAtom();
    maybeParseImportAttributes();
  ***REMOVED***
  semicolon();
***REMOVED***

function shouldParseExportStar() {
  if (isFlowEnabled) {
    return flowShouldParseExportStar();
  ***REMOVED*** else {
    return match(tt.star);
  ***REMOVED***
***REMOVED***

function parseExportStar() {
  if (isFlowEnabled) {
    flowParseExportStar();
  ***REMOVED*** else {
    baseParseExportStar();
  ***REMOVED***
***REMOVED***

export function baseParseExportStar() {
  expect(tt.star);

  if (isContextual(ContextualKeyword._as)) {
    parseExportNamespace();
  ***REMOVED*** else {
    parseExportFrom();
  ***REMOVED***
***REMOVED***

function parseExportNamespace() {
  next();
  state.tokens[state.tokens.length - 1].type = tt._as;
  parseIdentifier();
  parseExportSpecifiersMaybe();
  parseExportFrom();
***REMOVED***

function shouldParseExportDeclaration() {
  return (
    (isTypeScriptEnabled && tsIsDeclarationStart()) ||
    (isFlowEnabled && flowShouldParseExportDeclaration()) ||
    state.type === tt._var ||
    state.type === tt._const ||
    state.type === tt._let ||
    state.type === tt._function ||
    state.type === tt._class ||
    isContextual(ContextualKeyword._async) ||
    match(tt.at)
  );
***REMOVED***

// Parses a comma-separated list of module exports.
export function parseExportSpecifiers() {
  let first = true;

  // export { x, y as z ***REMOVED*** [from '...']
  expect(tt.braceL);

  while (!eat(tt.braceR) && !state.error) {
    if (first) {
      first = false;
    ***REMOVED*** else {
      expect(tt.comma);
      if (eat(tt.braceR)) {
        break;
      ***REMOVED***
    ***REMOVED***
    parseExportSpecifier();
  ***REMOVED***
***REMOVED***

function parseExportSpecifier() {
  if (isTypeScriptEnabled) {
    tsParseExportSpecifier();
    return;
  ***REMOVED***
  parseIdentifier();
  state.tokens[state.tokens.length - 1].identifierRole = IdentifierRole.ExportAccess;
  if (eatContextual(ContextualKeyword._as)) {
    parseIdentifier();
  ***REMOVED***
***REMOVED***

/**
 * Starting at the `module` token in an import, determine if it was truly an
 * import reflection token or just looks like one.
 *
 * Returns true for:
 * import module foo from "foo";
 * import module from from "foo";
 *
 * Returns false for:
 * import module from "foo";
 * import module, {bar***REMOVED*** from "foo";
 */
function isImportReflection() {
  const snapshot = state.snapshot();
  expectContextual(ContextualKeyword._module);
  if (eatContextual(ContextualKeyword._from)) {
    if (isContextual(ContextualKeyword._from)) {
      state.restoreFromSnapshot(snapshot);
      return true;
    ***REMOVED*** else {
      state.restoreFromSnapshot(snapshot);
      return false;
    ***REMOVED***
  ***REMOVED*** else if (match(tt.comma)) {
    state.restoreFromSnapshot(snapshot);
    return false;
  ***REMOVED*** else {
    state.restoreFromSnapshot(snapshot);
    return true;
  ***REMOVED***
***REMOVED***

/**
 * Eat the "module" token from the import reflection proposal.
 * https://github.com/tc39/proposal-import-reflection
 */
function parseMaybeImportReflection() {
  // isImportReflection does snapshot/restore, so only run it if we see the word
  // "module".
  if (isContextual(ContextualKeyword._module) && isImportReflection()) {
    next();
  ***REMOVED***
***REMOVED***

// Parses import declaration.

export function parseImport() {
  if (isTypeScriptEnabled && match(tt.name) && lookaheadType() === tt.eq) {
    tsParseImportEqualsDeclaration();
    return;
  ***REMOVED***
  if (isTypeScriptEnabled && isContextual(ContextualKeyword._type)) {
    const lookahead = lookaheadTypeAndKeyword();
    if (lookahead.type === tt.name && lookahead.contextualKeyword !== ContextualKeyword._from) {
      // One of these `import type` cases:
      // import type T = require('T');
      // import type A from 'A';
      expectContextual(ContextualKeyword._type);
      if (lookaheadType() === tt.eq) {
        tsParseImportEqualsDeclaration();
        return;
      ***REMOVED***
      // If this is an `import type...from` statement, then we already ate the
      // type token, so proceed to the regular import parser.
    ***REMOVED*** else if (lookahead.type === tt.star || lookahead.type === tt.braceL) {
      // One of these `import type` cases, in which case we can eat the type token
      // and proceed as normal:
      // import type * as A from 'A';
      // import type {a***REMOVED*** from 'A';
      expectContextual(ContextualKeyword._type);
    ***REMOVED***
    // Otherwise, we are importing the name "type".
  ***REMOVED***

  // import '...'
  if (match(tt.string)) {
    parseExprAtom();
  ***REMOVED*** else {
    parseMaybeImportReflection();
    parseImportSpecifiers();
    expectContextual(ContextualKeyword._from);
    parseExprAtom();
  ***REMOVED***
  maybeParseImportAttributes();
  semicolon();
***REMOVED***

// eslint-disable-next-line no-unused-vars
function shouldParseDefaultImport() {
  return match(tt.name);
***REMOVED***

function parseImportSpecifierLocal() {
  parseImportedIdentifier();
***REMOVED***

// Parses a comma-separated list of module imports.
function parseImportSpecifiers() {
  if (isFlowEnabled) {
    flowStartParseImportSpecifiers();
  ***REMOVED***

  let first = true;
  if (shouldParseDefaultImport()) {
    // import defaultObj, { x, y as z ***REMOVED*** from '...'
    parseImportSpecifierLocal();

    if (!eat(tt.comma)) return;
  ***REMOVED***

  if (match(tt.star)) {
    next();
    expectContextual(ContextualKeyword._as);

    parseImportSpecifierLocal();

    return;
  ***REMOVED***

  expect(tt.braceL);
  while (!eat(tt.braceR) && !state.error) {
    if (first) {
      first = false;
    ***REMOVED*** else {
      // Detect an attempt to deep destructure
      if (eat(tt.colon)) {
        unexpected(
          "ES2015 named imports do not destructure. Use another statement for destructuring after the import.",
        );
      ***REMOVED***

      expect(tt.comma);
      if (eat(tt.braceR)) {
        break;
      ***REMOVED***
    ***REMOVED***

    parseImportSpecifier();
  ***REMOVED***
***REMOVED***

function parseImportSpecifier() {
  if (isTypeScriptEnabled) {
    tsParseImportSpecifier();
    return;
  ***REMOVED***
  if (isFlowEnabled) {
    flowParseImportSpecifier();
    return;
  ***REMOVED***
  parseImportedIdentifier();
  if (isContextual(ContextualKeyword._as)) {
    state.tokens[state.tokens.length - 1].identifierRole = IdentifierRole.ImportAccess;
    next();
    parseImportedIdentifier();
  ***REMOVED***
***REMOVED***

/**
 * Parse import attributes like `with {type: "json"***REMOVED***`, or the legacy form
 * `assert {type: "json"***REMOVED***`.
 *
 * Import attributes technically have their own syntax, but are always parseable
 * as a plain JS object, so just do that for simplicity.
 */
function maybeParseImportAttributes() {
  if (match(tt._with) || (isContextual(ContextualKeyword._assert) && !hasPrecedingLineBreak())) {
    next();
    parseObj(false, false);
  ***REMOVED***
***REMOVED***
