// webpack/lib/config/defaults.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const fs = require("fs");
const path = require("path");
const {
	JAVASCRIPT_MODULE_TYPE_AUTO,
	JAVASCRIPT_MODULE_TYPE_ESM,
	JAVASCRIPT_MODULE_TYPE_DYNAMIC,
	JSON_MODULE_TYPE,
	WEBASSEMBLY_MODULE_TYPE_ASYNC,
	WEBASSEMBLY_MODULE_TYPE_SYNC,
	ASSET_MODULE_TYPE,
	ASSET_MODULE_TYPE_INLINE,
	ASSET_MODULE_TYPE_RESOURCE,
	CSS_MODULE_TYPE_AUTO,
	CSS_MODULE_TYPE,
	CSS_MODULE_TYPE_MODULE,
	CSS_MODULE_TYPE_GLOBAL
***REMOVED*** = require("../ModuleTypeConstants");
const Template = require("../Template");
const { cleverMerge ***REMOVED*** = require("../util/cleverMerge");
const {
	getTargetsProperties,
	getTargetProperties,
	getDefaultTarget
***REMOVED*** = require("./target");

/** @typedef {import("../../declarations/WebpackOptions").CacheOptions***REMOVED*** CacheOptions */
/** @typedef {import("../../declarations/WebpackOptions").CacheOptionsNormalized***REMOVED*** CacheOptionsNormalized */
/** @typedef {import("../../declarations/WebpackOptions").Context***REMOVED*** Context */
/** @typedef {import("../../declarations/WebpackOptions").CssGeneratorOptions***REMOVED*** CssGeneratorOptions */
/** @typedef {import("../../declarations/WebpackOptions").CssParserOptions***REMOVED*** CssParserOptions */
/** @typedef {import("../../declarations/WebpackOptions").EntryDescription***REMOVED*** EntryDescription */
/** @typedef {import("../../declarations/WebpackOptions").EntryNormalized***REMOVED*** Entry */
/** @typedef {import("../../declarations/WebpackOptions").EntryStaticNormalized***REMOVED*** EntryStaticNormalized */
/** @typedef {import("../../declarations/WebpackOptions").Environment***REMOVED*** Environment */
/** @typedef {import("../../declarations/WebpackOptions").Experiments***REMOVED*** Experiments */
/** @typedef {import("../../declarations/WebpackOptions").ExperimentsNormalized***REMOVED*** ExperimentsNormalized */
/** @typedef {import("../../declarations/WebpackOptions").ExternalsPresets***REMOVED*** ExternalsPresets */
/** @typedef {import("../../declarations/WebpackOptions").ExternalsType***REMOVED*** ExternalsType */
/** @typedef {import("../../declarations/WebpackOptions").FileCacheOptions***REMOVED*** FileCacheOptions */
/** @typedef {import("../../declarations/WebpackOptions").GeneratorOptionsByModuleTypeKnown***REMOVED*** GeneratorOptionsByModuleTypeKnown */
/** @typedef {import("../../declarations/WebpackOptions").InfrastructureLogging***REMOVED*** InfrastructureLogging */
/** @typedef {import("../../declarations/WebpackOptions").JavascriptParserOptions***REMOVED*** JavascriptParserOptions */
/** @typedef {import("../../declarations/WebpackOptions").JsonGeneratorOptions***REMOVED*** JsonGeneratorOptions */
/** @typedef {import("../../declarations/WebpackOptions").Library***REMOVED*** Library */
/** @typedef {import("../../declarations/WebpackOptions").LibraryName***REMOVED*** LibraryName */
/** @typedef {import("../../declarations/WebpackOptions").LibraryOptions***REMOVED*** LibraryOptions */
/** @typedef {import("../../declarations/WebpackOptions").LibraryType***REMOVED*** LibraryType */
/** @typedef {import("../../declarations/WebpackOptions").Loader***REMOVED*** Loader */
/** @typedef {import("../../declarations/WebpackOptions").Mode***REMOVED*** Mode */
/** @typedef {import("../../declarations/WebpackOptions").ModuleOptionsNormalized***REMOVED*** ModuleOptions */
/** @typedef {import("../../declarations/WebpackOptions").Node***REMOVED*** WebpackNode */
/** @typedef {import("../../declarations/WebpackOptions").Optimization***REMOVED*** Optimization */
/** @typedef {import("../../declarations/WebpackOptions").OptimizationSplitChunksOptions***REMOVED*** OptimizationSplitChunksOptions */
/** @typedef {import("../../declarations/WebpackOptions").OutputNormalized***REMOVED*** Output */
/** @typedef {import("../../declarations/WebpackOptions").ParserOptionsByModuleTypeKnown***REMOVED*** ParserOptionsByModuleTypeKnown */
/** @typedef {import("../../declarations/WebpackOptions").Performance***REMOVED*** Performance */
/** @typedef {import("../../declarations/WebpackOptions").ResolveOptions***REMOVED*** ResolveOptions */
/** @typedef {import("../../declarations/WebpackOptions").RuleSetRules***REMOVED*** RuleSetRules */
/** @typedef {import("../../declarations/WebpackOptions").SnapshotOptions***REMOVED*** SnapshotOptions */
/** @typedef {import("../../declarations/WebpackOptions").Target***REMOVED*** Target */
/** @typedef {import("../../declarations/WebpackOptions").WebpackOptions***REMOVED*** WebpackOptions */
/** @typedef {import("../../declarations/WebpackOptions").WebpackOptionsNormalized***REMOVED*** WebpackOptionsNormalized */
/** @typedef {import("../Compiler")***REMOVED*** Compiler */
/** @typedef {import("../Module")***REMOVED*** Module */
/** @typedef {import("./target").PlatformTargetProperties***REMOVED*** PlatformTargetProperties */
/** @typedef {import("./target").TargetProperties***REMOVED*** TargetProperties */

/**
 * @typedef {object***REMOVED*** ResolvedOptions
 * @property {PlatformTargetProperties | false***REMOVED*** platform - platform target properties
 */

const NODE_MODULES_REGEXP = /[\\/]node_modules[\\/]/i;
const DEFAULT_CACHE_NAME = "default";
const DEFAULTS = {
	// TODO webpack 6 - use xxhash64
	HASH_FUNCTION: "md4"
***REMOVED***;

/**
 * Sets a constant default value when undefined
 * @template T
 * @template {keyof T***REMOVED*** P
 * @param {T***REMOVED*** obj an object
 * @param {P***REMOVED*** prop a property of this object
 * @param {T[P]***REMOVED*** value a default value of the property
 * @returns {void***REMOVED***
 */
const D = (obj, prop, value) => {
	if (obj[prop] === undefined) {
		obj[prop] = value;
	***REMOVED***
***REMOVED***;

/**
 * Sets a dynamic default value when undefined, by calling the factory function
 * @template T
 * @template {keyof T***REMOVED*** P
 * @param {T***REMOVED*** obj an object
 * @param {P***REMOVED*** prop a property of this object
 * @param {() => T[P]***REMOVED*** factory a default value factory for the property
 * @returns {void***REMOVED***
 */
const F = (obj, prop, factory) => {
	if (obj[prop] === undefined) {
		obj[prop] = factory();
	***REMOVED***
***REMOVED***;

/**
 * Sets a dynamic default value when undefined, by calling the factory function.
 * factory must return an array or undefined
 * When the current value is already an array an contains "..." it's replaced with
 * the result of the factory function
 * @template T
 * @template {keyof T***REMOVED*** P
 * @param {T***REMOVED*** obj an object
 * @param {P***REMOVED*** prop a property of this object
 * @param {() => T[P]***REMOVED*** factory a default value factory for the property
 * @returns {void***REMOVED***
 */
const A = (obj, prop, factory) => {
	const value = obj[prop];
	if (value === undefined) {
		obj[prop] = factory();
	***REMOVED*** else if (Array.isArray(value)) {
		/** @type {EXPECTED_ANY[] | undefined***REMOVED*** */
		let newArray;
		for (let i = 0; i < value.length; i++) {
			const item = value[i];
			if (item === "...") {
				if (newArray === undefined) {
					newArray = value.slice(0, i);
					obj[prop] = /** @type {T[P]***REMOVED*** */ (/** @type {unknown***REMOVED*** */ (newArray));
				***REMOVED***
				const items = /** @type {EXPECTED_ANY[]***REMOVED*** */ (
					/** @type {unknown***REMOVED*** */ (factory())
				);
				if (items !== undefined) {
					for (const item of items) {
						newArray.push(item);
					***REMOVED***
				***REMOVED***
			***REMOVED*** else if (newArray !== undefined) {
				newArray.push(item);
			***REMOVED***
		***REMOVED***
	***REMOVED***
***REMOVED***;

/**
 * @param {WebpackOptionsNormalized***REMOVED*** options options to be modified
 * @returns {void***REMOVED***
 */
const applyWebpackOptionsBaseDefaults = options => {
	F(options, "context", () => process.cwd());
	applyInfrastructureLoggingDefaults(options.infrastructureLogging);
***REMOVED***;

/**
 * @param {WebpackOptionsNormalized***REMOVED*** options options to be modified
 * @param {number=***REMOVED*** compilerIndex index of compiler
 * @returns {ResolvedOptions***REMOVED*** Resolved options after apply defaults
 */
const applyWebpackOptionsDefaults = (options, compilerIndex) => {
	F(options, "context", () => process.cwd());
	F(options, "target", () =>
		getDefaultTarget(/** @type {string***REMOVED*** */ (options.context))
	);

	const { mode, name, target ***REMOVED*** = options;

	const targetProperties =
		target === false
			? /** @type {false***REMOVED*** */ (false)
			: typeof target === "string"
				? getTargetProperties(target, /** @type {Context***REMOVED*** */ (options.context))
				: getTargetsProperties(
						/** @type {string[]***REMOVED*** */ (target),
						/** @type {Context***REMOVED*** */ (options.context)
					);

	const development = mode === "development";
	const production = mode === "production" || !mode;

	if (typeof options.entry !== "function") {
		for (const key of Object.keys(options.entry)) {
			F(
				options.entry[key],
				"import",
				() => /** @type {[string]***REMOVED*** */ (["./src"])
			);
		***REMOVED***
	***REMOVED***

	F(options, "devtool", () => (development ? "eval" : false));
	D(options, "watch", false);
	D(options, "profile", false);
	D(options, "parallelism", 100);
	D(options, "recordsInputPath", false);
	D(options, "recordsOutputPath", false);

	applyExperimentsDefaults(options.experiments, {
		production,
		development,
		targetProperties
	***REMOVED***);

	const futureDefaults =
		/** @type {NonNullable<ExperimentsNormalized["futureDefaults"]>***REMOVED*** */
		(options.experiments.futureDefaults);

	F(options, "cache", () =>
		development ? { type: /** @type {"memory"***REMOVED*** */ ("memory") ***REMOVED*** : false
	);
	applyCacheDefaults(options.cache, {
		name: name || DEFAULT_CACHE_NAME,
		mode: mode || "production",
		development,
		cacheUnaffected: options.experiments.cacheUnaffected,
		futureDefaults,
		compilerIndex
	***REMOVED***);
	const cache = Boolean(options.cache);

	applySnapshotDefaults(options.snapshot, {
		production,
		futureDefaults
	***REMOVED***);

	applyOutputDefaults(options.output, {
		context: /** @type {Context***REMOVED*** */ (options.context),
		targetProperties,
		isAffectedByBrowserslist:
			target === undefined ||
			(typeof target === "string" && target.startsWith("browserslist")) ||
			(Array.isArray(target) &&
				target.some(target => target.startsWith("browserslist"))),
		outputModule:
			/** @type {NonNullable<ExperimentsNormalized["outputModule"]>***REMOVED*** */
			(options.experiments.outputModule),
		development,
		entry: options.entry,
		futureDefaults,
		asyncWebAssembly:
			/** @type {NonNullable<ExperimentsNormalized["asyncWebAssembly"]>***REMOVED*** */
			(options.experiments.asyncWebAssembly)
	***REMOVED***);

	applyModuleDefaults(options.module, {
		cache,
		syncWebAssembly:
			/** @type {NonNullable<ExperimentsNormalized["syncWebAssembly"]>***REMOVED*** */
			(options.experiments.syncWebAssembly),
		asyncWebAssembly:
			/** @type {NonNullable<ExperimentsNormalized["asyncWebAssembly"]>***REMOVED*** */
			(options.experiments.asyncWebAssembly),
		css:
			/** @type {NonNullable<ExperimentsNormalized["css"]>***REMOVED*** */
			(options.experiments.css),
		futureDefaults,
		isNode: targetProperties && targetProperties.node === true,
		uniqueName: /** @type {string***REMOVED*** */ (options.output.uniqueName),
		targetProperties,
		mode: options.mode
	***REMOVED***);

	applyExternalsPresetsDefaults(options.externalsPresets, {
		targetProperties,
		buildHttp: Boolean(options.experiments.buildHttp)
	***REMOVED***);

	applyLoaderDefaults(
		/** @type {NonNullable<WebpackOptionsNormalized["loader"]>***REMOVED*** */ (
			options.loader
		),
		{ targetProperties, environment: options.output.environment ***REMOVED***
	);

	F(options, "externalsType", () => {
		const validExternalTypes = require("../../schemas/WebpackOptions.json")
			.definitions.ExternalsType.enum;
		return options.output.library &&
			validExternalTypes.includes(options.output.library.type)
			? /** @type {ExternalsType***REMOVED*** */ (options.output.library.type)
			: options.output.module
				? "module-import"
				: "var";
	***REMOVED***);

	applyNodeDefaults(options.node, {
		futureDefaults:
			/** @type {NonNullable<WebpackOptionsNormalized["experiments"]["futureDefaults"]>***REMOVED*** */
			(options.experiments.futureDefaults),
		outputModule:
			/** @type {NonNullable<WebpackOptionsNormalized["output"]["module"]>***REMOVED*** */
			(options.output.module),
		targetProperties
	***REMOVED***);

	F(options, "performance", () =>
		production &&
		targetProperties &&
		(targetProperties.browser || targetProperties.browser === null)
			? {***REMOVED***
			: false
	);
	applyPerformanceDefaults(
		/** @type {NonNullable<WebpackOptionsNormalized["performance"]>***REMOVED*** */
		(options.performance),
		{
			production
		***REMOVED***
	);

	applyOptimizationDefaults(options.optimization, {
		development,
		production,
		css:
			/** @type {NonNullable<ExperimentsNormalized["css"]>***REMOVED*** */
			(options.experiments.css),
		records: Boolean(options.recordsInputPath || options.recordsOutputPath)
	***REMOVED***);

	options.resolve = cleverMerge(
		getResolveDefaults({
			cache,
			context: /** @type {Context***REMOVED*** */ (options.context),
			targetProperties,
			mode: /** @type {Mode***REMOVED*** */ (options.mode),
			css:
				/** @type {NonNullable<ExperimentsNormalized["css"]>***REMOVED*** */
				(options.experiments.css)
		***REMOVED***),
		options.resolve
	);

	options.resolveLoader = cleverMerge(
		getResolveLoaderDefaults({ cache ***REMOVED***),
		options.resolveLoader
	);

	return {
		platform:
			targetProperties === false
				? targetProperties
				: {
						web: targetProperties.web,
						browser: targetProperties.browser,
						webworker: targetProperties.webworker,
						node: targetProperties.node,
						nwjs: targetProperties.nwjs,
						electron: targetProperties.electron
					***REMOVED***
	***REMOVED***;
***REMOVED***;

/**
 * @param {ExperimentsNormalized***REMOVED*** experiments options
 * @param {object***REMOVED*** options options
 * @param {boolean***REMOVED*** options.production is production
 * @param {boolean***REMOVED*** options.development is development mode
 * @param {TargetProperties | false***REMOVED*** options.targetProperties target properties
 * @returns {void***REMOVED***
 */
const applyExperimentsDefaults = (
	experiments,
	{ production, development, targetProperties ***REMOVED***
) => {
	D(experiments, "futureDefaults", false);
	D(experiments, "backCompat", !experiments.futureDefaults);
	D(experiments, "syncWebAssembly", false);
	D(experiments, "asyncWebAssembly", experiments.futureDefaults);
	D(experiments, "outputModule", false);
	D(experiments, "layers", false);
	D(experiments, "lazyCompilation", undefined);
	D(experiments, "buildHttp", undefined);
	D(experiments, "cacheUnaffected", experiments.futureDefaults);
	F(experiments, "css", () => (experiments.futureDefaults ? true : undefined));

	// TODO webpack 6: remove this. topLevelAwait should be enabled by default
	let shouldEnableTopLevelAwait = true;
	if (typeof experiments.topLevelAwait === "boolean") {
		shouldEnableTopLevelAwait = experiments.topLevelAwait;
	***REMOVED***
	D(experiments, "topLevelAwait", shouldEnableTopLevelAwait);

	if (typeof experiments.buildHttp === "object") {
		D(experiments.buildHttp, "frozen", production);
		D(experiments.buildHttp, "upgrade", false);
	***REMOVED***
***REMOVED***;

/**
 * @param {CacheOptionsNormalized***REMOVED*** cache options
 * @param {object***REMOVED*** options options
 * @param {string***REMOVED*** options.name name
 * @param {Mode***REMOVED*** options.mode mode
 * @param {boolean***REMOVED*** options.futureDefaults is future defaults enabled
 * @param {boolean***REMOVED*** options.development is development mode
 * @param {number=***REMOVED*** options.compilerIndex index of compiler
 * @param {Experiments["cacheUnaffected"]***REMOVED*** options.cacheUnaffected the cacheUnaffected experiment is enabled
 * @returns {void***REMOVED***
 */
const applyCacheDefaults = (
	cache,
	{ name, mode, development, cacheUnaffected, compilerIndex, futureDefaults ***REMOVED***
) => {
	if (cache === false) return;
	switch (cache.type) {
		case "filesystem":
			F(cache, "name", () =>
				compilerIndex !== undefined
					? `${`${name***REMOVED***-${mode***REMOVED***`***REMOVED***__compiler${compilerIndex + 1***REMOVED***__`
					: `${name***REMOVED***-${mode***REMOVED***`
			);
			D(cache, "version", "");
			F(cache, "cacheDirectory", () => {
				const cwd = process.cwd();
				/** @type {string | undefined***REMOVED*** */
				let dir = cwd;
				for (;;) {
					try {
						if (fs.statSync(path.join(dir, "package.json")).isFile()) break;
						// eslint-disable-next-line no-empty
					***REMOVED*** catch (_err) {***REMOVED***
					const parent = path.dirname(dir);
					if (dir === parent) {
						dir = undefined;
						break;
					***REMOVED***
					dir = parent;
				***REMOVED***
				if (!dir) {
					return path.resolve(cwd, ".cache/webpack");
				***REMOVED*** else if (process.versions.pnp === "1") {
					return path.resolve(dir, ".pnp/.cache/webpack");
				***REMOVED*** else if (process.versions.pnp === "3") {
					return path.resolve(dir, ".yarn/.cache/webpack");
				***REMOVED***
				return path.resolve(dir, "node_modules/.cache/webpack");
			***REMOVED***);
			F(cache, "cacheLocation", () =>
				path.resolve(
					/** @type {NonNullable<FileCacheOptions["cacheDirectory"]>***REMOVED*** */
					(cache.cacheDirectory),
					/** @type {NonNullable<FileCacheOptions["name"]>***REMOVED*** */ (cache.name)
				)
			);
			D(cache, "hashAlgorithm", futureDefaults ? "xxhash64" : "md4");
			D(cache, "store", "pack");
			D(cache, "compression", false);
			D(cache, "profile", false);
			D(cache, "idleTimeout", 60000);
			D(cache, "idleTimeoutForInitialStore", 5000);
			D(cache, "idleTimeoutAfterLargeChanges", 1000);
			D(cache, "maxMemoryGenerations", development ? 5 : Infinity);
			D(cache, "maxAge", 1000 * 60 * 60 * 24 * 60); // 1 month
			D(cache, "allowCollectingMemory", development);
			D(cache, "memoryCacheUnaffected", development && cacheUnaffected);
			D(cache, "readonly", false);
			D(
				/** @type {NonNullable<FileCacheOptions["buildDependencies"]>***REMOVED*** */
				(cache.buildDependencies),
				"defaultWebpack",
				[path.resolve(__dirname, "..") + path.sep]
			);
			break;
		case "memory":
			D(cache, "maxGenerations", Infinity);
			D(cache, "cacheUnaffected", development && cacheUnaffected);
			break;
	***REMOVED***
***REMOVED***;

/**
 * @param {SnapshotOptions***REMOVED*** snapshot options
 * @param {object***REMOVED*** options options
 * @param {boolean***REMOVED*** options.production is production
 * @param {boolean***REMOVED*** options.futureDefaults is future defaults enabled
 * @returns {void***REMOVED***
 */
const applySnapshotDefaults = (snapshot, { production, futureDefaults ***REMOVED***) => {
	if (futureDefaults) {
		F(snapshot, "managedPaths", () =>
			process.versions.pnp === "3"
				? [
						/^(.+?(?:[\\/]\.yarn[\\/]unplugged[\\/][^\\/]+)?[\\/]node_modules[\\/])/
					]
				: [/^(.+?[\\/]node_modules[\\/])/]
		);
		F(snapshot, "immutablePaths", () =>
			process.versions.pnp === "3"
				? [/^(.+?[\\/]cache[\\/][^\\/]+\.zip[\\/]node_modules[\\/])/]
				: []
		);
	***REMOVED*** else {
		A(snapshot, "managedPaths", () => {
			if (process.versions.pnp === "3") {
				const match =
					/^(.+?)[\\/]cache[\\/]watchpack-npm-[^\\/]+\.zip[\\/]node_modules[\\/]/.exec(
						require.resolve("watchpack")
					);
				if (match) {
					return [path.resolve(match[1], "unplugged")];
				***REMOVED***
			***REMOVED*** else {
				const match = /^(.+?[\\/]node_modules[\\/])/.exec(
					require.resolve("watchpack")
				);
				if (match) {
					return [match[1]];
				***REMOVED***
			***REMOVED***
			return [];
		***REMOVED***);
		A(snapshot, "immutablePaths", () => {
			if (process.versions.pnp === "1") {
				const match =
					/^(.+?[\\/]v4)[\\/]npm-watchpack-[^\\/]+-[\da-f]{40***REMOVED***[\\/]node_modules[\\/]/.exec(
						require.resolve("watchpack")
					);
				if (match) {
					return [match[1]];
				***REMOVED***
			***REMOVED*** else if (process.versions.pnp === "3") {
				const match =
					/^(.+?)[\\/]watchpack-npm-[^\\/]+\.zip[\\/]node_modules[\\/]/.exec(
						require.resolve("watchpack")
					);
				if (match) {
					return [match[1]];
				***REMOVED***
			***REMOVED***
			return [];
		***REMOVED***);
	***REMOVED***
	F(snapshot, "unmanagedPaths", () => []);
	F(snapshot, "resolveBuildDependencies", () => ({
		timestamp: true,
		hash: true
	***REMOVED***));
	F(snapshot, "buildDependencies", () => ({ timestamp: true, hash: true ***REMOVED***));
	F(snapshot, "module", () =>
		production ? { timestamp: true, hash: true ***REMOVED*** : { timestamp: true ***REMOVED***
	);
	F(snapshot, "resolve", () =>
		production ? { timestamp: true, hash: true ***REMOVED*** : { timestamp: true ***REMOVED***
	);
***REMOVED***;

/**
 * @param {JavascriptParserOptions***REMOVED*** parserOptions parser options
 * @param {object***REMOVED*** options options
 * @param {boolean***REMOVED*** options.futureDefaults is future defaults enabled
 * @param {boolean***REMOVED*** options.isNode is node target platform
 * @returns {void***REMOVED***
 */
const applyJavascriptParserOptionsDefaults = (
	parserOptions,
	{ futureDefaults, isNode ***REMOVED***
) => {
	D(parserOptions, "unknownContextRequest", ".");
	D(parserOptions, "unknownContextRegExp", false);
	D(parserOptions, "unknownContextRecursive", true);
	D(parserOptions, "unknownContextCritical", true);
	D(parserOptions, "exprContextRequest", ".");
	D(parserOptions, "exprContextRegExp", false);
	D(parserOptions, "exprContextRecursive", true);
	D(parserOptions, "exprContextCritical", true);
	D(parserOptions, "wrappedContextRegExp", /.*/);
	D(parserOptions, "wrappedContextRecursive", true);
	D(parserOptions, "wrappedContextCritical", false);
	D(parserOptions, "strictThisContextOnImports", false);
	D(parserOptions, "importMeta", true);
	D(parserOptions, "dynamicImportMode", "lazy");
	D(parserOptions, "dynamicImportPrefetch", false);
	D(parserOptions, "dynamicImportPreload", false);
	D(parserOptions, "dynamicImportFetchPriority", false);
	D(parserOptions, "createRequire", isNode);
	if (futureDefaults) D(parserOptions, "exportsPresence", "error");
***REMOVED***;

/**
 * @param {JsonGeneratorOptions***REMOVED*** generatorOptions generator options
 * @returns {void***REMOVED***
 */
const applyJsonGeneratorOptionsDefaults = generatorOptions => {
	D(generatorOptions, "JSONParse", true);
***REMOVED***;

/**
 * @param {CssGeneratorOptions***REMOVED*** generatorOptions generator options
 * @param {object***REMOVED*** options options
 * @param {TargetProperties | false***REMOVED*** options.targetProperties target properties
 * @returns {void***REMOVED***
 */
const applyCssGeneratorOptionsDefaults = (
	generatorOptions,
	{ targetProperties ***REMOVED***
) => {
	D(
		generatorOptions,
		"exportsOnly",
		!targetProperties || targetProperties.document === false
	);
	D(generatorOptions, "esModule", true);
***REMOVED***;

/**
 * @param {ModuleOptions***REMOVED*** module options
 * @param {object***REMOVED*** options options
 * @param {boolean***REMOVED*** options.cache is caching enabled
 * @param {boolean***REMOVED*** options.syncWebAssembly is syncWebAssembly enabled
 * @param {boolean***REMOVED*** options.asyncWebAssembly is asyncWebAssembly enabled
 * @param {boolean***REMOVED*** options.css is css enabled
 * @param {boolean***REMOVED*** options.futureDefaults is future defaults enabled
 * @param {string***REMOVED*** options.uniqueName the unique name
 * @param {boolean***REMOVED*** options.isNode is node target platform
 * @param {TargetProperties | false***REMOVED*** options.targetProperties target properties
 * @param {Mode | undefined***REMOVED*** options.mode mode
 * @returns {void***REMOVED***
 */
const applyModuleDefaults = (
	module,
	{
		cache,
		syncWebAssembly,
		asyncWebAssembly,
		css,
		futureDefaults,
		isNode,
		uniqueName,
		targetProperties,
		mode
	***REMOVED***
) => {
	if (cache) {
		D(
			module,
			"unsafeCache",
			/**
			 * @param {Module***REMOVED*** module module
			 * @returns {boolean***REMOVED*** true, if we want to cache the module
			 */
			module => {
				const name = module.nameForCondition();
				if (!name) {
					return false;
				***REMOVED***
				return NODE_MODULES_REGEXP.test(name);
			***REMOVED***
		);
	***REMOVED*** else {
		D(module, "unsafeCache", false);
	***REMOVED***

	F(module.parser, ASSET_MODULE_TYPE, () => ({***REMOVED***));
	F(
		/** @type {NonNullable<ParserOptionsByModuleTypeKnown[ASSET_MODULE_TYPE]>***REMOVED*** */
		(module.parser[ASSET_MODULE_TYPE]),
		"dataUrlCondition",
		() => ({***REMOVED***)
	);
	if (
		typeof (
			/** @type {NonNullable<ParserOptionsByModuleTypeKnown[ASSET_MODULE_TYPE]>***REMOVED*** */
			(module.parser[ASSET_MODULE_TYPE]).dataUrlCondition
		) === "object"
	) {
		D(
			/** @type {NonNullable<ParserOptionsByModuleTypeKnown[ASSET_MODULE_TYPE]>***REMOVED*** */
			(module.parser[ASSET_MODULE_TYPE]).dataUrlCondition,
			"maxSize",
			8096
		);
	***REMOVED***

	F(module.parser, "javascript", () => ({***REMOVED***));
	F(module.parser, JSON_MODULE_TYPE, () => ({***REMOVED***));
	D(
		/** @type {NonNullable<ParserOptionsByModuleTypeKnown[JSON_MODULE_TYPE]>***REMOVED*** */
		(module.parser[JSON_MODULE_TYPE]),
		"exportsDepth",
		mode === "development" ? 1 : Infinity
	);

	applyJavascriptParserOptionsDefaults(
		/** @type {NonNullable<ParserOptionsByModuleTypeKnown["javascript"]>***REMOVED*** */
		(module.parser.javascript),
		{
			futureDefaults,
			isNode
		***REMOVED***
	);

	F(module.generator, "json", () => ({***REMOVED***));
	applyJsonGeneratorOptionsDefaults(
		/** @type {NonNullable<GeneratorOptionsByModuleTypeKnown["json"]>***REMOVED*** */
		(module.generator.json)
	);

	if (css) {
		F(module.parser, CSS_MODULE_TYPE, () => ({***REMOVED***));

		D(
			/** @type {NonNullable<ParserOptionsByModuleTypeKnown[CSS_MODULE_TYPE]>***REMOVED*** */
			(module.parser[CSS_MODULE_TYPE]),
			"import",
			true
		);
		D(
			/** @type {NonNullable<ParserOptionsByModuleTypeKnown[CSS_MODULE_TYPE]>***REMOVED*** */
			(module.parser[CSS_MODULE_TYPE]),
			"url",
			true
		);
		D(
			/** @type {NonNullable<ParserOptionsByModuleTypeKnown[CSS_MODULE_TYPE]>***REMOVED*** */
			(module.parser[CSS_MODULE_TYPE]),
			"namedExports",
			true
		);

		F(module.generator, CSS_MODULE_TYPE, () => ({***REMOVED***));

		applyCssGeneratorOptionsDefaults(
			/** @type {NonNullable<GeneratorOptionsByModuleTypeKnown[CSS_MODULE_TYPE]>***REMOVED*** */
			(module.generator[CSS_MODULE_TYPE]),
			{ targetProperties ***REMOVED***
		);

		const localIdentName =
			uniqueName.length > 0 ? "[uniqueName]-[id]-[local]" : "[id]-[local]";

		F(module.generator, CSS_MODULE_TYPE_AUTO, () => ({***REMOVED***));
		D(
			/** @type {NonNullable<GeneratorOptionsByModuleTypeKnown[CSS_MODULE_TYPE_AUTO]>***REMOVED*** */
			(module.generator[CSS_MODULE_TYPE_AUTO]),
			"localIdentName",
			localIdentName
		);
		D(
			/** @type {NonNullable<GeneratorOptionsByModuleTypeKnown[CSS_MODULE_TYPE_AUTO]>***REMOVED*** */
			(module.generator[CSS_MODULE_TYPE_AUTO]),
			"exportsConvention",
			"as-is"
		);

		F(module.generator, CSS_MODULE_TYPE_MODULE, () => ({***REMOVED***));
		D(
			/** @type {NonNullable<GeneratorOptionsByModuleTypeKnown[CSS_MODULE_TYPE_MODULE]>***REMOVED*** */
			(module.generator[CSS_MODULE_TYPE_MODULE]),
			"localIdentName",
			localIdentName
		);
		D(
			/** @type {NonNullable<GeneratorOptionsByModuleTypeKnown[CSS_MODULE_TYPE_MODULE]>***REMOVED*** */
			(module.generator[CSS_MODULE_TYPE_MODULE]),
			"exportsConvention",
			"as-is"
		);

		F(module.generator, CSS_MODULE_TYPE_GLOBAL, () => ({***REMOVED***));
		D(
			/** @type {NonNullable<GeneratorOptionsByModuleTypeKnown[CSS_MODULE_TYPE_GLOBAL]>***REMOVED*** */
			(module.generator[CSS_MODULE_TYPE_GLOBAL]),
			"localIdentName",
			localIdentName
		);
		D(
			/** @type {NonNullable<GeneratorOptionsByModuleTypeKnown[CSS_MODULE_TYPE_GLOBAL]>***REMOVED*** */
			(module.generator[CSS_MODULE_TYPE_GLOBAL]),
			"exportsConvention",
			"as-is"
		);
	***REMOVED***

	A(module, "defaultRules", () => {
		const esm = {
			type: JAVASCRIPT_MODULE_TYPE_ESM,
			resolve: {
				byDependency: {
					esm: {
						fullySpecified: true
					***REMOVED***
				***REMOVED***
			***REMOVED***
		***REMOVED***;
		const commonjs = {
			type: JAVASCRIPT_MODULE_TYPE_DYNAMIC
		***REMOVED***;
		/** @type {RuleSetRules***REMOVED*** */
		const rules = [
			{
				mimetype: "application/node",
				type: JAVASCRIPT_MODULE_TYPE_AUTO
			***REMOVED***,
			{
				test: /\.json$/i,
				type: JSON_MODULE_TYPE
			***REMOVED***,
			{
				mimetype: "application/json",
				type: JSON_MODULE_TYPE
			***REMOVED***,
			{
				test: /\.mjs$/i,
				...esm
			***REMOVED***,
			{
				test: /\.js$/i,
				descriptionData: {
					type: "module"
				***REMOVED***,
				...esm
			***REMOVED***,
			{
				test: /\.cjs$/i,
				...commonjs
			***REMOVED***,
			{
				test: /\.js$/i,
				descriptionData: {
					type: "commonjs"
				***REMOVED***,
				...commonjs
			***REMOVED***,
			{
				mimetype: {
					or: ["text/javascript", "application/javascript"]
				***REMOVED***,
				...esm
			***REMOVED***
		];
		if (asyncWebAssembly) {
			const wasm = {
				type: WEBASSEMBLY_MODULE_TYPE_ASYNC,
				rules: [
					{
						descriptionData: {
							type: "module"
						***REMOVED***,
						resolve: {
							fullySpecified: true
						***REMOVED***
					***REMOVED***
				]
			***REMOVED***;
			rules.push({
				test: /\.wasm$/i,
				...wasm
			***REMOVED***);
			rules.push({
				mimetype: "application/wasm",
				...wasm
			***REMOVED***);
		***REMOVED*** else if (syncWebAssembly) {
			const wasm = {
				type: WEBASSEMBLY_MODULE_TYPE_SYNC,
				rules: [
					{
						descriptionData: {
							type: "module"
						***REMOVED***,
						resolve: {
							fullySpecified: true
						***REMOVED***
					***REMOVED***
				]
			***REMOVED***;
			rules.push({
				test: /\.wasm$/i,
				...wasm
			***REMOVED***);
			rules.push({
				mimetype: "application/wasm",
				...wasm
			***REMOVED***);
		***REMOVED***
		if (css) {
			const resolve = {
				fullySpecified: true,
				preferRelative: true
			***REMOVED***;
			rules.push({
				test: /\.css$/i,
				type: CSS_MODULE_TYPE_AUTO,
				resolve
			***REMOVED***);
			rules.push({
				mimetype: "text/css+module",
				type: CSS_MODULE_TYPE_MODULE,
				resolve
			***REMOVED***);
			rules.push({
				mimetype: "text/css",
				type: CSS_MODULE_TYPE,
				resolve
			***REMOVED***);
		***REMOVED***
		rules.push(
			{
				dependency: "url",
				oneOf: [
					{
						scheme: /^data$/,
						type: ASSET_MODULE_TYPE_INLINE
					***REMOVED***,
					{
						type: ASSET_MODULE_TYPE_RESOURCE
					***REMOVED***
				]
			***REMOVED***,
			{
				assert: { type: JSON_MODULE_TYPE ***REMOVED***,
				type: JSON_MODULE_TYPE
			***REMOVED***,
			{
				with: { type: JSON_MODULE_TYPE ***REMOVED***,
				type: JSON_MODULE_TYPE
			***REMOVED***
		);
		return rules;
	***REMOVED***);
***REMOVED***;

/**
 * @param {Output***REMOVED*** output options
 * @param {object***REMOVED*** options options
 * @param {string***REMOVED*** options.context context
 * @param {TargetProperties | false***REMOVED*** options.targetProperties target properties
 * @param {boolean***REMOVED*** options.isAffectedByBrowserslist is affected by browserslist
 * @param {boolean***REMOVED*** options.outputModule is outputModule experiment enabled
 * @param {boolean***REMOVED*** options.development is development mode
 * @param {Entry***REMOVED*** options.entry entry option
 * @param {boolean***REMOVED*** options.futureDefaults is future defaults enabled
 * @param {boolean***REMOVED*** options.asyncWebAssembly is asyncWebAssembly enabled
 * @returns {void***REMOVED***
 */
const applyOutputDefaults = (
	output,
	{
		context,
		targetProperties: tp,
		isAffectedByBrowserslist,
		outputModule,
		development,
		entry,
		futureDefaults,
		asyncWebAssembly
	***REMOVED***
) => {
	/**
	 * @param {Library=***REMOVED*** library the library option
	 * @returns {string***REMOVED*** a readable library name
	 */
	const getLibraryName = library => {
		const libraryName =
			typeof library === "object" &&
			library &&
			!Array.isArray(library) &&
			"type" in library
				? library.name
				: /** @type {LibraryName***REMOVED*** */ (library);
		if (Array.isArray(libraryName)) {
			return libraryName.join(".");
		***REMOVED*** else if (typeof libraryName === "object") {
			return getLibraryName(libraryName.root);
		***REMOVED*** else if (typeof libraryName === "string") {
			return libraryName;
		***REMOVED***
		return "";
	***REMOVED***;

	F(output, "uniqueName", () => {
		const libraryName = getLibraryName(output.library).replace(
			/^\[(\\*[\w:]+\\*)\](\.)|(\.)\[(\\*[\w:]+\\*)\](?=\.|$)|\[(\\*[\w:]+\\*)\]/g,
			(m, a, d1, d2, b, c) => {
				const content = a || b || c;
				return content.startsWith("\\") && content.endsWith("\\")
					? `${d2 || ""***REMOVED***[${content.slice(1, -1)***REMOVED***]${d1 || ""***REMOVED***`
					: "";
			***REMOVED***
		);
		if (libraryName) return libraryName;
		const pkgPath = path.resolve(context, "package.json");
		try {
			const packageInfo = JSON.parse(fs.readFileSync(pkgPath, "utf-8"));
			return packageInfo.name || "";
		***REMOVED*** catch (err) {
			if (/** @type {Error & { code: string ***REMOVED******REMOVED*** */ (err).code !== "ENOENT") {
				/** @type {Error & { code: string ***REMOVED******REMOVED*** */
				(err).message +=
					`\nwhile determining default 'output.uniqueName' from 'name' in ${pkgPath***REMOVED***`;
				throw err;
			***REMOVED***
			return "";
		***REMOVED***
	***REMOVED***);

	F(output, "module", () => Boolean(outputModule));

	const environment = /** @type {Environment***REMOVED*** */ (output.environment);
	/**
	 * @param {boolean | undefined***REMOVED*** v value
	 * @returns {boolean***REMOVED*** true, when v is truthy or undefined
	 */
	const optimistic = v => v || v === undefined;
	/**
	 * @param {boolean | undefined***REMOVED*** v value
	 * @param {boolean | undefined***REMOVED*** c condition
	 * @returns {boolean | undefined***REMOVED*** true, when v is truthy or undefined, or c is truthy
	 */
	const conditionallyOptimistic = (v, c) => (v === undefined && c) || v;

	F(
		environment,
		"globalThis",
		() => /** @type {boolean | undefined***REMOVED*** */ (tp && tp.globalThis)
	);
	F(
		environment,
		"bigIntLiteral",
		() =>
			tp && optimistic(/** @type {boolean | undefined***REMOVED*** */ (tp.bigIntLiteral))
	);
	F(
		environment,
		"const",
		() => tp && optimistic(/** @type {boolean | undefined***REMOVED*** */ (tp.const))
	);
	F(
		environment,
		"arrowFunction",
		() =>
			tp && optimistic(/** @type {boolean | undefined***REMOVED*** */ (tp.arrowFunction))
	);
	F(
		environment,
		"asyncFunction",
		() =>
			tp && optimistic(/** @type {boolean | undefined***REMOVED*** */ (tp.asyncFunction))
	);
	F(
		environment,
		"forOf",
		() => tp && optimistic(/** @type {boolean | undefined***REMOVED*** */ (tp.forOf))
	);
	F(
		environment,
		"destructuring",
		() =>
			tp && optimistic(/** @type {boolean | undefined***REMOVED*** */ (tp.destructuring))
	);
	F(
		environment,
		"optionalChaining",
		() =>
			tp && optimistic(/** @type {boolean | undefined***REMOVED*** */ (tp.optionalChaining))
	);
	F(
		environment,
		"nodePrefixForCoreModules",
		() =>
			tp &&
			optimistic(
				/** @type {boolean | undefined***REMOVED*** */ (tp.nodePrefixForCoreModules)
			)
	);
	F(
		environment,
		"templateLiteral",
		() =>
			tp && optimistic(/** @type {boolean | undefined***REMOVED*** */ (tp.templateLiteral))
	);
	F(environment, "dynamicImport", () =>
		conditionallyOptimistic(
			/** @type {boolean | undefined***REMOVED*** */ (tp && tp.dynamicImport),
			output.module
		)
	);
	F(environment, "dynamicImportInWorker", () =>
		conditionallyOptimistic(
			/** @type {boolean | undefined***REMOVED*** */ (tp && tp.dynamicImportInWorker),
			output.module
		)
	);
	F(environment, "module", () =>
		conditionallyOptimistic(
			/** @type {boolean | undefined***REMOVED*** */ (tp && tp.module),
			output.module
		)
	);
	F(
		environment,
		"document",
		() => tp && optimistic(/** @type {boolean | undefined***REMOVED*** */ (tp.document))
	);

	D(output, "filename", output.module ? "[name].mjs" : "[name].js");
	F(output, "iife", () => !output.module);
	D(output, "importFunctionName", "import");
	D(output, "importMetaName", "import.meta");
	F(output, "chunkFilename", () => {
		const filename =
			/** @type {NonNullable<Output["chunkFilename"]>***REMOVED*** */
			(output.filename);
		if (typeof filename !== "function") {
			const hasName = filename.includes("[name]");
			const hasId = filename.includes("[id]");
			const hasChunkHash = filename.includes("[chunkhash]");
			const hasContentHash = filename.includes("[contenthash]");
			// Anything changing depending on chunk is fine
			if (hasChunkHash || hasContentHash || hasName || hasId) return filename;
			// Otherwise prefix "[id]." in front of the basename to make it changing
			return filename.replace(/(^|\/)([^/]*(?:\?|$))/, "$1[id].$2");
		***REMOVED***
		return output.module ? "[id].mjs" : "[id].js";
	***REMOVED***);
	F(output, "cssFilename", () => {
		const filename =
			/** @type {NonNullable<Output["cssFilename"]>***REMOVED*** */
			(output.filename);
		if (typeof filename !== "function") {
			return filename.replace(/\.[mc]?js(\?|$)/, ".css$1");
		***REMOVED***
		return "[id].css";
	***REMOVED***);
	F(output, "cssChunkFilename", () => {
		const chunkFilename =
			/** @type {NonNullable<Output["cssChunkFilename"]>***REMOVED*** */
			(output.chunkFilename);
		if (typeof chunkFilename !== "function") {
			return chunkFilename.replace(/\.[mc]?js(\?|$)/, ".css$1");
		***REMOVED***
		return "[id].css";
	***REMOVED***);
	D(output, "assetModuleFilename", "[hash][ext][query]");
	D(output, "webassemblyModuleFilename", "[hash].module.wasm");
	D(output, "compareBeforeEmit", true);
	D(output, "charset", !futureDefaults);
	const uniqueNameId = Template.toIdentifier(
		/** @type {NonNullable<Output["uniqueName"]>***REMOVED*** */ (output.uniqueName)
	);
	F(output, "hotUpdateGlobal", () => `webpackHotUpdate${uniqueNameId***REMOVED***`);
	F(output, "chunkLoadingGlobal", () => `webpackChunk${uniqueNameId***REMOVED***`);
	F(output, "globalObject", () => {
		if (tp) {
			if (tp.global) return "global";
			if (tp.globalThis) return "globalThis";
		***REMOVED***
		return "self";
	***REMOVED***);
	F(output, "chunkFormat", () => {
		if (tp) {
			const helpMessage = isAffectedByBrowserslist
				? "Make sure that your 'browserslist' includes only platforms that support these features or select an appropriate 'target' to allow selecting a chunk format by default. Alternatively specify the 'output.chunkFormat' directly."
				: "Select an appropriate 'target' to allow selecting one by default, or specify the 'output.chunkFormat' directly.";
			if (output.module) {
				if (environment.dynamicImport) return "module";
				if (tp.document) return "array-push";
				throw new Error(
					"For the selected environment is no default ESM chunk format available:\n" +
						"ESM exports can be chosen when 'import()' is available.\n" +
						`JSONP Array push can be chosen when 'document' is available.\n${
							helpMessage
						***REMOVED***`
				);
			***REMOVED*** else {
				if (tp.document) return "array-push";
				if (tp.require) return "commonjs";
				if (tp.nodeBuiltins) return "commonjs";
				if (tp.importScripts) return "array-push";
				throw new Error(
					"For the selected environment is no default script chunk format available:\n" +
						"JSONP Array push can be chosen when 'document' or 'importScripts' is available.\n" +
						`CommonJs exports can be chosen when 'require' or node builtins are available.\n${
							helpMessage
						***REMOVED***`
				);
			***REMOVED***
		***REMOVED***
		throw new Error(
			"Chunk format can't be selected by default when no target is specified"
		);
	***REMOVED***);
	D(output, "asyncChunks", true);
	F(output, "chunkLoading", () => {
		if (tp) {
			switch (output.chunkFormat) {
				case "array-push":
					if (tp.document) return "jsonp";
					if (tp.importScripts) return "import-scripts";
					break;
				case "commonjs":
					if (tp.require) return "require";
					if (tp.nodeBuiltins) return "async-node";
					break;
				case "module":
					if (environment.dynamicImport) return "import";
					break;
			***REMOVED***
			if (
				(tp.require === null ||
					tp.nodeBuiltins === null ||
					tp.document === null ||
					tp.importScripts === null) &&
				output.module &&
				environment.dynamicImport
			) {
				return "universal";
			***REMOVED***
		***REMOVED***
		return false;
	***REMOVED***);
	F(output, "workerChunkLoading", () => {
		if (tp) {
			switch (output.chunkFormat) {
				case "array-push":
					if (tp.importScriptsInWorker) return "import-scripts";
					break;
				case "commonjs":
					if (tp.require) return "require";
					if (tp.nodeBuiltins) return "async-node";
					break;
				case "module":
					if (environment.dynamicImportInWorker) return "import";
					break;
			***REMOVED***
			if (
				(tp.require === null ||
					tp.nodeBuiltins === null ||
					tp.importScriptsInWorker === null) &&
				output.module &&
				environment.dynamicImport
			) {
				return "universal";
			***REMOVED***
		***REMOVED***
		return false;
	***REMOVED***);
	F(output, "wasmLoading", () => {
		if (tp) {
			if (tp.fetchWasm) return "fetch";
			if (tp.nodeBuiltins) return "async-node";
			if (
				(tp.nodeBuiltins === null || tp.fetchWasm === null) &&
				asyncWebAssembly &&
				output.module &&
				environment.dynamicImport
			) {
				return "universal";
			***REMOVED***
		***REMOVED***
		return false;
	***REMOVED***);
	F(output, "workerWasmLoading", () => output.wasmLoading);
	F(output, "devtoolNamespace", () => output.uniqueName);
	if (output.library) {
		F(output.library, "type", () => (output.module ? "module" : "var"));
	***REMOVED***
	F(output, "path", () => path.join(process.cwd(), "dist"));
	F(output, "pathinfo", () => development);
	D(output, "sourceMapFilename", "[file].map[query]");
	D(
		output,
		"hotUpdateChunkFilename",
		`[id].[fullhash].hot-update.${output.module ? "mjs" : "js"***REMOVED***`
	);
	D(output, "hotUpdateMainFilename", "[runtime].[fullhash].hot-update.json");
	D(output, "crossOriginLoading", false);
	F(output, "scriptType", () => (output.module ? "module" : false));
	D(
		output,
		"publicPath",
		(tp && (tp.document || tp.importScripts)) || output.scriptType === "module"
			? "auto"
			: ""
	);
	D(output, "workerPublicPath", "");
	D(output, "chunkLoadTimeout", 120000);
	F(output, "hashFunction", () => {
		if (futureDefaults) {
			DEFAULTS.HASH_FUNCTION = "xxhash64";
			return "xxhash64";
		***REMOVED***

		return "md4";
	***REMOVED***);
	D(output, "hashDigest", "hex");
	D(output, "hashDigestLength", futureDefaults ? 16 : 20);
	D(output, "strictModuleErrorHandling", false);
	D(output, "strictModuleExceptionHandling", false);

	const { trustedTypes ***REMOVED*** = output;
	if (trustedTypes) {
		F(
			trustedTypes,
			"policyName",
			() =>
				/** @type {NonNullable<Output["uniqueName"]>***REMOVED*** */
				(output.uniqueName).replace(/[^a-zA-Z0-9\-#=_/@.%]+/g, "_") || "webpack"
		);
		D(trustedTypes, "onPolicyCreationFailure", "stop");
	***REMOVED***

	/**
	 * @param {(entryDescription: EntryDescription) => void***REMOVED*** fn iterator
	 * @returns {void***REMOVED***
	 */
	const forEachEntry = fn => {
		for (const name of Object.keys(entry)) {
			fn(/** @type {{[k: string] : EntryDescription***REMOVED******REMOVED*** */ (entry)[name]);
		***REMOVED***
	***REMOVED***;
	A(output, "enabledLibraryTypes", () => {
		/** @type {LibraryType[]***REMOVED*** */
		const enabledLibraryTypes = [];
		if (output.library) {
			enabledLibraryTypes.push(output.library.type);
		***REMOVED***
		forEachEntry(desc => {
			if (desc.library) {
				enabledLibraryTypes.push(desc.library.type);
			***REMOVED***
		***REMOVED***);
		return enabledLibraryTypes;
	***REMOVED***);

	A(output, "enabledChunkLoadingTypes", () => {
		const enabledChunkLoadingTypes = new Set();
		if (output.chunkLoading) {
			enabledChunkLoadingTypes.add(output.chunkLoading);
		***REMOVED***
		if (output.workerChunkLoading) {
			enabledChunkLoadingTypes.add(output.workerChunkLoading);
		***REMOVED***
		forEachEntry(desc => {
			if (desc.chunkLoading) {
				enabledChunkLoadingTypes.add(desc.chunkLoading);
			***REMOVED***
		***REMOVED***);
		return Array.from(enabledChunkLoadingTypes);
	***REMOVED***);

	A(output, "enabledWasmLoadingTypes", () => {
		const enabledWasmLoadingTypes = new Set();
		if (output.wasmLoading) {
			enabledWasmLoadingTypes.add(output.wasmLoading);
		***REMOVED***
		if (output.workerWasmLoading) {
			enabledWasmLoadingTypes.add(output.workerWasmLoading);
		***REMOVED***
		forEachEntry(desc => {
			if (desc.wasmLoading) {
				enabledWasmLoadingTypes.add(desc.wasmLoading);
			***REMOVED***
		***REMOVED***);
		return Array.from(enabledWasmLoadingTypes);
	***REMOVED***);
***REMOVED***;

/**
 * @param {ExternalsPresets***REMOVED*** externalsPresets options
 * @param {object***REMOVED*** options options
 * @param {TargetProperties | false***REMOVED*** options.targetProperties target properties
 * @param {boolean***REMOVED*** options.buildHttp buildHttp experiment enabled
 * @returns {void***REMOVED***
 */
const applyExternalsPresetsDefaults = (
	externalsPresets,
	{ targetProperties, buildHttp ***REMOVED***
) => {
	D(
		externalsPresets,
		"web",
		/** @type {boolean | undefined***REMOVED*** */
		(!buildHttp && targetProperties && targetProperties.web)
	);
	D(
		externalsPresets,
		"node",
		/** @type {boolean | undefined***REMOVED*** */
		(targetProperties && targetProperties.node)
	);
	D(
		externalsPresets,
		"nwjs",
		/** @type {boolean | undefined***REMOVED*** */
		(targetProperties && targetProperties.nwjs)
	);
	D(
		externalsPresets,
		"electron",
		/** @type {boolean | undefined***REMOVED*** */
		(targetProperties && targetProperties.electron)
	);
	D(
		externalsPresets,
		"electronMain",
		/** @type {boolean | undefined***REMOVED*** */
		(
			targetProperties &&
				targetProperties.electron &&
				targetProperties.electronMain
		)
	);
	D(
		externalsPresets,
		"electronPreload",
		/** @type {boolean | undefined***REMOVED*** */
		(
			targetProperties &&
				targetProperties.electron &&
				targetProperties.electronPreload
		)
	);
	D(
		externalsPresets,
		"electronRenderer",
		/** @type {boolean | undefined***REMOVED*** */
		(
			targetProperties &&
				targetProperties.electron &&
				targetProperties.electronRenderer
		)
	);
***REMOVED***;

/**
 * @param {Loader***REMOVED*** loader options
 * @param {object***REMOVED*** options options
 * @param {TargetProperties | false***REMOVED*** options.targetProperties target properties
 * @param {Environment***REMOVED*** options.environment environment
 * @returns {void***REMOVED***
 */
const applyLoaderDefaults = (loader, { targetProperties, environment ***REMOVED***) => {
	F(loader, "target", () => {
		if (targetProperties) {
			if (targetProperties.electron) {
				if (targetProperties.electronMain) return "electron-main";
				if (targetProperties.electronPreload) return "electron-preload";
				if (targetProperties.electronRenderer) return "electron-renderer";
				return "electron";
			***REMOVED***
			if (targetProperties.nwjs) return "nwjs";
			if (targetProperties.node) return "node";
			if (targetProperties.web) return "web";
		***REMOVED***
	***REMOVED***);
	D(loader, "environment", environment);
***REMOVED***;

/**
 * @param {WebpackNode***REMOVED*** node options
 * @param {object***REMOVED*** options options
 * @param {TargetProperties | false***REMOVED*** options.targetProperties target properties
 * @param {boolean***REMOVED*** options.futureDefaults is future defaults enabled
 * @param {boolean***REMOVED*** options.outputModule is output type is module
 * @returns {void***REMOVED***
 */
const applyNodeDefaults = (
	node,
	{ futureDefaults, outputModule, targetProperties ***REMOVED***
) => {
	if (node === false) return;

	F(node, "global", () => {
		if (targetProperties && targetProperties.global) return false;
		// TODO webpack 6 should always default to false
		return futureDefaults ? "warn" : true;
	***REMOVED***);

	const handlerForNames = () => {
		if (targetProperties && targetProperties.node)
			return outputModule ? "node-module" : "eval-only";
		// TODO webpack 6 should always default to false
		return futureDefaults ? "warn-mock" : "mock";
	***REMOVED***;

	F(node, "__filename", handlerForNames);
	F(node, "__dirname", handlerForNames);
***REMOVED***;

/**
 * @param {Performance***REMOVED*** performance options
 * @param {object***REMOVED*** options options
 * @param {boolean***REMOVED*** options.production is production
 * @returns {void***REMOVED***
 */
const applyPerformanceDefaults = (performance, { production ***REMOVED***) => {
	if (performance === false) return;
	D(performance, "maxAssetSize", 250000);
	D(performance, "maxEntrypointSize", 250000);
	F(performance, "hints", () => (production ? "warning" : false));
***REMOVED***;

/**
 * @param {Optimization***REMOVED*** optimization options
 * @param {object***REMOVED*** options options
 * @param {boolean***REMOVED*** options.production is production
 * @param {boolean***REMOVED*** options.development is development
 * @param {boolean***REMOVED*** options.css is css enabled
 * @param {boolean***REMOVED*** options.records using records
 * @returns {void***REMOVED***
 */
const applyOptimizationDefaults = (
	optimization,
	{ production, development, css, records ***REMOVED***
) => {
	D(optimization, "removeAvailableModules", false);
	D(optimization, "removeEmptyChunks", true);
	D(optimization, "mergeDuplicateChunks", true);
	D(optimization, "flagIncludedChunks", production);
	F(optimization, "moduleIds", () => {
		if (production) return "deterministic";
		if (development) return "named";
		return "natural";
	***REMOVED***);
	F(optimization, "chunkIds", () => {
		if (production) return "deterministic";
		if (development) return "named";
		return "natural";
	***REMOVED***);
	F(optimization, "sideEffects", () => (production ? true : "flag"));
	D(optimization, "providedExports", true);
	D(optimization, "usedExports", production);
	D(optimization, "innerGraph", production);
	D(optimization, "mangleExports", production);
	D(optimization, "concatenateModules", production);
	D(optimization, "avoidEntryIife", production);
	D(optimization, "runtimeChunk", false);
	D(optimization, "emitOnErrors", !production);
	D(optimization, "checkWasmTypes", production);
	D(optimization, "mangleWasmImports", false);
	D(optimization, "portableRecords", records);
	D(optimization, "realContentHash", production);
	D(optimization, "minimize", production);
	A(optimization, "minimizer", () => [
		{
			apply: compiler => {
				// Lazy load the Terser plugin
				const TerserPlugin = require("terser-webpack-plugin");
				new TerserPlugin({
					terserOptions: {
						compress: {
							passes: 2
						***REMOVED***
					***REMOVED***
				***REMOVED***).apply(/** @type {EXPECTED_ANY***REMOVED*** */ (compiler));
			***REMOVED***
		***REMOVED***
	]);
	F(optimization, "nodeEnv", () => {
		if (production) return "production";
		if (development) return "development";
		return false;
	***REMOVED***);
	const { splitChunks ***REMOVED*** = optimization;
	if (splitChunks) {
		A(splitChunks, "defaultSizeTypes", () =>
			css ? ["javascript", "css", "unknown"] : ["javascript", "unknown"]
		);
		D(splitChunks, "hidePathInfo", production);
		D(splitChunks, "chunks", "async");
		D(splitChunks, "usedExports", optimization.usedExports === true);
		D(splitChunks, "minChunks", 1);
		F(splitChunks, "minSize", () => (production ? 20000 : 10000));
		F(splitChunks, "minRemainingSize", () => (development ? 0 : undefined));
		F(splitChunks, "enforceSizeThreshold", () => (production ? 50000 : 30000));
		F(splitChunks, "maxAsyncRequests", () => (production ? 30 : Infinity));
		F(splitChunks, "maxInitialRequests", () => (production ? 30 : Infinity));
		D(splitChunks, "automaticNameDelimiter", "-");
		const cacheGroups =
			/** @type {NonNullable<OptimizationSplitChunksOptions["cacheGroups"]>***REMOVED*** */
			(splitChunks.cacheGroups);
		F(cacheGroups, "default", () => ({
			idHint: "",
			reuseExistingChunk: true,
			minChunks: 2,
			priority: -20
		***REMOVED***));
		F(cacheGroups, "defaultVendors", () => ({
			idHint: "vendors",
			reuseExistingChunk: true,
			test: NODE_MODULES_REGEXP,
			priority: -10
		***REMOVED***));
	***REMOVED***
***REMOVED***;

/**
 * @param {object***REMOVED*** options options
 * @param {boolean***REMOVED*** options.cache is cache enable
 * @param {string***REMOVED*** options.context build context
 * @param {TargetProperties | false***REMOVED*** options.targetProperties target properties
 * @param {Mode***REMOVED*** options.mode mode
 * @param {boolean***REMOVED*** options.css is css enabled
 * @returns {ResolveOptions***REMOVED*** resolve options
 */
const getResolveDefaults = ({
	cache,
	context,
	targetProperties,
	mode,
	css
***REMOVED***) => {
	/** @type {string[]***REMOVED*** */
	const conditions = ["webpack"];

	conditions.push(mode === "development" ? "development" : "production");

	if (targetProperties) {
		if (targetProperties.webworker) conditions.push("worker");
		if (targetProperties.node) conditions.push("node");
		if (targetProperties.web) conditions.push("browser");
		if (targetProperties.electron) conditions.push("electron");
		if (targetProperties.nwjs) conditions.push("nwjs");
	***REMOVED***

	const jsExtensions = [".js", ".json", ".wasm"];

	const tp = targetProperties;
	const browserField =
		tp && tp.web && (!tp.node || (tp.electron && tp.electronRenderer));

	/** @type {() => ResolveOptions***REMOVED*** */
	const cjsDeps = () => ({
		aliasFields: browserField ? ["browser"] : [],
		mainFields: browserField ? ["browser", "module", "..."] : ["module", "..."],
		conditionNames: ["require", "module", "..."],
		extensions: [...jsExtensions]
	***REMOVED***);
	/** @type {() => ResolveOptions***REMOVED*** */
	const esmDeps = () => ({
		aliasFields: browserField ? ["browser"] : [],
		mainFields: browserField ? ["browser", "module", "..."] : ["module", "..."],
		conditionNames: ["import", "module", "..."],
		extensions: [...jsExtensions]
	***REMOVED***);

	/** @type {ResolveOptions***REMOVED*** */
	const resolveOptions = {
		cache,
		modules: ["node_modules"],
		conditionNames: conditions,
		mainFiles: ["index"],
		extensions: [],
		aliasFields: [],
		exportsFields: ["exports"],
		roots: [context],
		mainFields: ["main"],
		importsFields: ["imports"],
		byDependency: {
			wasm: esmDeps(),
			esm: esmDeps(),
			loaderImport: esmDeps(),
			url: {
				preferRelative: true
			***REMOVED***,
			worker: {
				...esmDeps(),
				preferRelative: true
			***REMOVED***,
			commonjs: cjsDeps(),
			amd: cjsDeps(),
			// for backward-compat: loadModule
			loader: cjsDeps(),
			// for backward-compat: Custom Dependency
			unknown: cjsDeps(),
			// for backward-compat: getResolve without dependencyType
			undefined: cjsDeps()
		***REMOVED***
	***REMOVED***;

	if (css) {
		const styleConditions = [];

		styleConditions.push("webpack");
		styleConditions.push(mode === "development" ? "development" : "production");
		styleConditions.push("style");

		/** @type {NonNullable<ResolveOptions["byDependency"]>***REMOVED*** */
		(resolveOptions.byDependency)["css-import"] = {
			// We avoid using any main files because we have to be consistent with CSS `@import`
			// and CSS `@import` does not handle `main` files in directories,
			// you should always specify the full URL for styles
			mainFiles: [],
			mainFields: ["style", "..."],
			conditionNames: styleConditions,
			extensions: [".css"],
			preferRelative: true
		***REMOVED***;
	***REMOVED***

	return resolveOptions;
***REMOVED***;

/**
 * @param {object***REMOVED*** options options
 * @param {boolean***REMOVED*** options.cache is cache enable
 * @returns {ResolveOptions***REMOVED*** resolve options
 */
const getResolveLoaderDefaults = ({ cache ***REMOVED***) => {
	/** @type {ResolveOptions***REMOVED*** */
	const resolveOptions = {
		cache,
		conditionNames: ["loader", "require", "node"],
		exportsFields: ["exports"],
		mainFields: ["loader", "main"],
		extensions: [".js"],
		mainFiles: ["index"]
	***REMOVED***;

	return resolveOptions;
***REMOVED***;

/**
 * @param {InfrastructureLogging***REMOVED*** infrastructureLogging options
 * @returns {void***REMOVED***
 */
const applyInfrastructureLoggingDefaults = infrastructureLogging => {
	F(infrastructureLogging, "stream", () => process.stderr);
	const tty =
		/** @type {NonNullable<InfrastructureLogging["stream"]>***REMOVED*** */
		(infrastructureLogging.stream).isTTY && process.env.TERM !== "dumb";
	D(infrastructureLogging, "level", "info");
	D(infrastructureLogging, "debug", false);
	D(infrastructureLogging, "colors", tty);
	D(infrastructureLogging, "appendOnly", !tty);
***REMOVED***;

module.exports.applyWebpackOptionsBaseDefaults =
	applyWebpackOptionsBaseDefaults;
module.exports.applyWebpackOptionsDefaults = applyWebpackOptionsDefaults;
module.exports.DEFAULTS = DEFAULTS;
