// object-inspect/index.js
var hasMap = typeof Map === 'function' && Map.prototype;
var mapSizeDescriptor = Object.getOwnPropertyDescriptor && hasMap ? Object.getOwnPropertyDescriptor(Map.prototype, 'size') : null;
var mapSize = hasMap && mapSizeDescriptor && typeof mapSizeDescriptor.get === 'function' ? mapSizeDescriptor.get : null;
var mapForEach = hasMap && Map.prototype.forEach;
var hasSet = typeof Set === 'function' && Set.prototype;
var setSizeDescriptor = Object.getOwnPropertyDescriptor && hasSet ? Object.getOwnPropertyDescriptor(Set.prototype, 'size') : null;
var setSize = hasSet && setSizeDescriptor && typeof setSizeDescriptor.get === 'function' ? setSizeDescriptor.get : null;
var setForEach = hasSet && Set.prototype.forEach;
var hasWeakMap = typeof WeakMap === 'function' && WeakMap.prototype;
var weakMapHas = hasWeakMap ? WeakMap.prototype.has : null;
var hasWeakSet = typeof WeakSet === 'function' && WeakSet.prototype;
var weakSetHas = hasWeakSet ? WeakSet.prototype.has : null;
var hasWeakRef = typeof WeakRef === 'function' && WeakRef.prototype;
var weakRefDeref = hasWeakRef ? WeakRef.prototype.deref : null;
var booleanValueOf = Boolean.prototype.valueOf;
var objectToString = Object.prototype.toString;
var functionToString = Function.prototype.toString;
var $match = String.prototype.match;
var $slice = String.prototype.slice;
var $replace = String.prototype.replace;
var $toUpperCase = String.prototype.toUpperCase;
var $toLowerCase = String.prototype.toLowerCase;
var $test = RegExp.prototype.test;
var $concat = Array.prototype.concat;
var $join = Array.prototype.join;
var $arrSlice = Array.prototype.slice;
var $floor = Math.floor;
var bigIntValueOf = typeof BigInt === 'function' ? BigInt.prototype.valueOf : null;
var gOPS = Object.getOwnPropertySymbols;
var symToString = typeof Symbol === 'function' && typeof Symbol.iterator === 'symbol' ? Symbol.prototype.toString : null;
var hasShammedSymbols = typeof Symbol === 'function' && typeof Symbol.iterator === 'object';
// ie, `has-tostringtag/shams
var toStringTag = typeof Symbol === 'function' && Symbol.toStringTag && (typeof Symbol.toStringTag === hasShammedSymbols ? 'object' : 'symbol')
    ? Symbol.toStringTag
    : null;
var isEnumerable = Object.prototype.propertyIsEnumerable;

var gPO = (typeof Reflect === 'function' ? Reflect.getPrototypeOf : Object.getPrototypeOf) || (
    [].__proto__ === Array.prototype // eslint-disable-line no-proto
        ? function (O) {
            return O.__proto__; // eslint-disable-line no-proto
        ***REMOVED***
        : null
);

function addNumericSeparator(num, str) {
    if (
        num === Infinity
        || num === -Infinity
        || num !== num
        || (num && num > -1000 && num < 1000)
        || $test.call(/e/, str)
    ) {
        return str;
    ***REMOVED***
    var sepRegex = /[0-9](?=(?:[0-9]{3***REMOVED***)+(?![0-9]))/g;
    if (typeof num === 'number') {
        var int = num < 0 ? -$floor(-num) : $floor(num); // trunc(num)
        if (int !== num) {
            var intStr = String(int);
            var dec = $slice.call(str, intStr.length + 1);
            return $replace.call(intStr, sepRegex, '$&_') + '.' + $replace.call($replace.call(dec, /([0-9]{3***REMOVED***)/g, '$&_'), /_$/, '');
        ***REMOVED***
    ***REMOVED***
    return $replace.call(str, sepRegex, '$&_');
***REMOVED***

var utilInspect = require('./util.inspect');
var inspectCustom = utilInspect.custom;
var inspectSymbol = isSymbol(inspectCustom) ? inspectCustom : null;

var quotes = {
    __proto__: null,
    'double': '"',
    single: "'"
***REMOVED***;
var quoteREs = {
    __proto__: null,
    'double': /(["\\])/g,
    single: /(['\\])/g
***REMOVED***;

module.exports = function inspect_(obj, options, depth, seen) {
    var opts = options || {***REMOVED***;

    if (has(opts, 'quoteStyle') && !has(quotes, opts.quoteStyle)) {
        throw new TypeError('option "quoteStyle" must be "single" or "double"');
    ***REMOVED***
    if (
        has(opts, 'maxStringLength') && (typeof opts.maxStringLength === 'number'
            ? opts.maxStringLength < 0 && opts.maxStringLength !== Infinity
            : opts.maxStringLength !== null
        )
    ) {
        throw new TypeError('option "maxStringLength", if provided, must be a positive integer, Infinity, or `null`');
    ***REMOVED***
    var customInspect = has(opts, 'customInspect') ? opts.customInspect : true;
    if (typeof customInspect !== 'boolean' && customInspect !== 'symbol') {
        throw new TypeError('option "customInspect", if provided, must be `true`, `false`, or `\'symbol\'`');
    ***REMOVED***

    if (
        has(opts, 'indent')
        && opts.indent !== null
        && opts.indent !== '\t'
        && !(parseInt(opts.indent, 10) === opts.indent && opts.indent > 0)
    ) {
        throw new TypeError('option "indent" must be "\\t", an integer > 0, or `null`');
    ***REMOVED***
    if (has(opts, 'numericSeparator') && typeof opts.numericSeparator !== 'boolean') {
        throw new TypeError('option "numericSeparator", if provided, must be `true` or `false`');
    ***REMOVED***
    var numericSeparator = opts.numericSeparator;

    if (typeof obj === 'undefined') {
        return 'undefined';
    ***REMOVED***
    if (obj === null) {
        return 'null';
    ***REMOVED***
    if (typeof obj === 'boolean') {
        return obj ? 'true' : 'false';
    ***REMOVED***

    if (typeof obj === 'string') {
        return inspectString(obj, opts);
    ***REMOVED***
    if (typeof obj === 'number') {
        if (obj === 0) {
            return Infinity / obj > 0 ? '0' : '-0';
        ***REMOVED***
        var str = String(obj);
        return numericSeparator ? addNumericSeparator(obj, str) : str;
    ***REMOVED***
    if (typeof obj === 'bigint') {
        var bigIntStr = String(obj) + 'n';
        return numericSeparator ? addNumericSeparator(obj, bigIntStr) : bigIntStr;
    ***REMOVED***

    var maxDepth = typeof opts.depth === 'undefined' ? 5 : opts.depth;
    if (typeof depth === 'undefined') { depth = 0; ***REMOVED***
    if (depth >= maxDepth && maxDepth > 0 && typeof obj === 'object') {
        return isArray(obj) ? '[Array]' : '[Object]';
    ***REMOVED***

    var indent = getIndent(opts, depth);

    if (typeof seen === 'undefined') {
        seen = [];
    ***REMOVED*** else if (indexOf(seen, obj) >= 0) {
        return '[Circular]';
    ***REMOVED***

    function inspect(value, from, noIndent) {
        if (from) {
            seen = $arrSlice.call(seen);
            seen.push(from);
        ***REMOVED***
        if (noIndent) {
            var newOpts = {
                depth: opts.depth
            ***REMOVED***;
            if (has(opts, 'quoteStyle')) {
                newOpts.quoteStyle = opts.quoteStyle;
            ***REMOVED***
            return inspect_(value, newOpts, depth + 1, seen);
        ***REMOVED***
        return inspect_(value, opts, depth + 1, seen);
    ***REMOVED***

    if (typeof obj === 'function' && !isRegExp(obj)) { // in older engines, regexes are callable
        var name = nameOf(obj);
        var keys = arrObjKeys(obj, inspect);
        return '[Function' + (name ? ': ' + name : ' (anonymous)') + ']' + (keys.length > 0 ? ' { ' + $join.call(keys, ', ') + ' ***REMOVED***' : '');
    ***REMOVED***
    if (isSymbol(obj)) {
        var symString = hasShammedSymbols ? $replace.call(String(obj), /^(Symbol\(.*\))_[^)]*$/, '$1') : symToString.call(obj);
        return typeof obj === 'object' && !hasShammedSymbols ? markBoxed(symString) : symString;
    ***REMOVED***
    if (isElement(obj)) {
        var s = '<' + $toLowerCase.call(String(obj.nodeName));
        var attrs = obj.attributes || [];
        for (var i = 0; i < attrs.length; i++) {
            s += ' ' + attrs[i].name + '=' + wrapQuotes(quote(attrs[i].value), 'double', opts);
        ***REMOVED***
        s += '>';
        if (obj.childNodes && obj.childNodes.length) { s += '...'; ***REMOVED***
        s += '</' + $toLowerCase.call(String(obj.nodeName)) + '>';
        return s;
    ***REMOVED***
    if (isArray(obj)) {
        if (obj.length === 0) { return '[]'; ***REMOVED***
        var xs = arrObjKeys(obj, inspect);
        if (indent && !singleLineValues(xs)) {
            return '[' + indentedJoin(xs, indent) + ']';
        ***REMOVED***
        return '[ ' + $join.call(xs, ', ') + ' ]';
    ***REMOVED***
    if (isError(obj)) {
        var parts = arrObjKeys(obj, inspect);
        if (!('cause' in Error.prototype) && 'cause' in obj && !isEnumerable.call(obj, 'cause')) {
            return '{ [' + String(obj) + '] ' + $join.call($concat.call('[cause]: ' + inspect(obj.cause), parts), ', ') + ' ***REMOVED***';
        ***REMOVED***
        if (parts.length === 0) { return '[' + String(obj) + ']'; ***REMOVED***
        return '{ [' + String(obj) + '] ' + $join.call(parts, ', ') + ' ***REMOVED***';
    ***REMOVED***
    if (typeof obj === 'object' && customInspect) {
        if (inspectSymbol && typeof obj[inspectSymbol] === 'function' && utilInspect) {
            return utilInspect(obj, { depth: maxDepth - depth ***REMOVED***);
        ***REMOVED*** else if (customInspect !== 'symbol' && typeof obj.inspect === 'function') {
            return obj.inspect();
        ***REMOVED***
    ***REMOVED***
    if (isMap(obj)) {
        var mapParts = [];
        if (mapForEach) {
            mapForEach.call(obj, function (value, key) {
                mapParts.push(inspect(key, obj, true) + ' => ' + inspect(value, obj));
            ***REMOVED***);
        ***REMOVED***
        return collectionOf('Map', mapSize.call(obj), mapParts, indent);
    ***REMOVED***
    if (isSet(obj)) {
        var setParts = [];
        if (setForEach) {
            setForEach.call(obj, function (value) {
                setParts.push(inspect(value, obj));
            ***REMOVED***);
        ***REMOVED***
        return collectionOf('Set', setSize.call(obj), setParts, indent);
    ***REMOVED***
    if (isWeakMap(obj)) {
        return weakCollectionOf('WeakMap');
    ***REMOVED***
    if (isWeakSet(obj)) {
        return weakCollectionOf('WeakSet');
    ***REMOVED***
    if (isWeakRef(obj)) {
        return weakCollectionOf('WeakRef');
    ***REMOVED***
    if (isNumber(obj)) {
        return markBoxed(inspect(Number(obj)));
    ***REMOVED***
    if (isBigInt(obj)) {
        return markBoxed(inspect(bigIntValueOf.call(obj)));
    ***REMOVED***
    if (isBoolean(obj)) {
        return markBoxed(booleanValueOf.call(obj));
    ***REMOVED***
    if (isString(obj)) {
        return markBoxed(inspect(String(obj)));
    ***REMOVED***
    // note: in IE 8, sometimes `global !== window` but both are the prototypes of each other
    /* eslint-env browser */
    if (typeof window !== 'undefined' && obj === window) {
        return '{ [object Window] ***REMOVED***';
    ***REMOVED***
    if (
        (typeof globalThis !== 'undefined' && obj === globalThis)
        || (typeof global !== 'undefined' && obj === global)
    ) {
        return '{ [object globalThis] ***REMOVED***';
    ***REMOVED***
    if (!isDate(obj) && !isRegExp(obj)) {
        var ys = arrObjKeys(obj, inspect);
        var isPlainObject = gPO ? gPO(obj) === Object.prototype : obj instanceof Object || obj.constructor === Object;
        var protoTag = obj instanceof Object ? '' : 'null prototype';
        var stringTag = !isPlainObject && toStringTag && Object(obj) === obj && toStringTag in obj ? $slice.call(toStr(obj), 8, -1) : protoTag ? 'Object' : '';
        var constructorTag = isPlainObject || typeof obj.constructor !== 'function' ? '' : obj.constructor.name ? obj.constructor.name + ' ' : '';
        var tag = constructorTag + (stringTag || protoTag ? '[' + $join.call($concat.call([], stringTag || [], protoTag || []), ': ') + '] ' : '');
        if (ys.length === 0) { return tag + '{***REMOVED***'; ***REMOVED***
        if (indent) {
            return tag + '{' + indentedJoin(ys, indent) + '***REMOVED***';
        ***REMOVED***
        return tag + '{ ' + $join.call(ys, ', ') + ' ***REMOVED***';
    ***REMOVED***
    return String(obj);
***REMOVED***;

function wrapQuotes(s, defaultStyle, opts) {
    var style = opts.quoteStyle || defaultStyle;
    var quoteChar = quotes[style];
    return quoteChar + s + quoteChar;
***REMOVED***

function quote(s) {
    return $replace.call(String(s), /"/g, '&quot;');
***REMOVED***

function canTrustToString(obj) {
    return !toStringTag || !(typeof obj === 'object' && (toStringTag in obj || typeof obj[toStringTag] !== 'undefined'));
***REMOVED***
function isArray(obj) { return toStr(obj) === '[object Array]' && canTrustToString(obj); ***REMOVED***
function isDate(obj) { return toStr(obj) === '[object Date]' && canTrustToString(obj); ***REMOVED***
function isRegExp(obj) { return toStr(obj) === '[object RegExp]' && canTrustToString(obj); ***REMOVED***
function isError(obj) { return toStr(obj) === '[object Error]' && canTrustToString(obj); ***REMOVED***
function isString(obj) { return toStr(obj) === '[object String]' && canTrustToString(obj); ***REMOVED***
function isNumber(obj) { return toStr(obj) === '[object Number]' && canTrustToString(obj); ***REMOVED***
function isBoolean(obj) { return toStr(obj) === '[object Boolean]' && canTrustToString(obj); ***REMOVED***

// Symbol and BigInt do have Symbol.toStringTag by spec, so that can't be used to eliminate false positives
function isSymbol(obj) {
    if (hasShammedSymbols) {
        return obj && typeof obj === 'object' && obj instanceof Symbol;
    ***REMOVED***
    if (typeof obj === 'symbol') {
        return true;
    ***REMOVED***
    if (!obj || typeof obj !== 'object' || !symToString) {
        return false;
    ***REMOVED***
    try {
        symToString.call(obj);
        return true;
    ***REMOVED*** catch (e) {***REMOVED***
    return false;
***REMOVED***

function isBigInt(obj) {
    if (!obj || typeof obj !== 'object' || !bigIntValueOf) {
        return false;
    ***REMOVED***
    try {
        bigIntValueOf.call(obj);
        return true;
    ***REMOVED*** catch (e) {***REMOVED***
    return false;
***REMOVED***

var hasOwn = Object.prototype.hasOwnProperty || function (key) { return key in this; ***REMOVED***;
function has(obj, key) {
    return hasOwn.call(obj, key);
***REMOVED***

function toStr(obj) {
    return objectToString.call(obj);
***REMOVED***

function nameOf(f) {
    if (f.name) { return f.name; ***REMOVED***
    var m = $match.call(functionToString.call(f), /^function\s*([\w$]+)/);
    if (m) { return m[1]; ***REMOVED***
    return null;
***REMOVED***

function indexOf(xs, x) {
    if (xs.indexOf) { return xs.indexOf(x); ***REMOVED***
    for (var i = 0, l = xs.length; i < l; i++) {
        if (xs[i] === x) { return i; ***REMOVED***
    ***REMOVED***
    return -1;
***REMOVED***

function isMap(x) {
    if (!mapSize || !x || typeof x !== 'object') {
        return false;
    ***REMOVED***
    try {
        mapSize.call(x);
        try {
            setSize.call(x);
        ***REMOVED*** catch (s) {
            return true;
        ***REMOVED***
        return x instanceof Map; // core-js workaround, pre-v2.5.0
    ***REMOVED*** catch (e) {***REMOVED***
    return false;
***REMOVED***

function isWeakMap(x) {
    if (!weakMapHas || !x || typeof x !== 'object') {
        return false;
    ***REMOVED***
    try {
        weakMapHas.call(x, weakMapHas);
        try {
            weakSetHas.call(x, weakSetHas);
        ***REMOVED*** catch (s) {
            return true;
        ***REMOVED***
        return x instanceof WeakMap; // core-js workaround, pre-v2.5.0
    ***REMOVED*** catch (e) {***REMOVED***
    return false;
***REMOVED***

function isWeakRef(x) {
    if (!weakRefDeref || !x || typeof x !== 'object') {
        return false;
    ***REMOVED***
    try {
        weakRefDeref.call(x);
        return true;
    ***REMOVED*** catch (e) {***REMOVED***
    return false;
***REMOVED***

function isSet(x) {
    if (!setSize || !x || typeof x !== 'object') {
        return false;
    ***REMOVED***
    try {
        setSize.call(x);
        try {
            mapSize.call(x);
        ***REMOVED*** catch (m) {
            return true;
        ***REMOVED***
        return x instanceof Set; // core-js workaround, pre-v2.5.0
    ***REMOVED*** catch (e) {***REMOVED***
    return false;
***REMOVED***

function isWeakSet(x) {
    if (!weakSetHas || !x || typeof x !== 'object') {
        return false;
    ***REMOVED***
    try {
        weakSetHas.call(x, weakSetHas);
        try {
            weakMapHas.call(x, weakMapHas);
        ***REMOVED*** catch (s) {
            return true;
        ***REMOVED***
        return x instanceof WeakSet; // core-js workaround, pre-v2.5.0
    ***REMOVED*** catch (e) {***REMOVED***
    return false;
***REMOVED***

function isElement(x) {
    if (!x || typeof x !== 'object') { return false; ***REMOVED***
    if (typeof HTMLElement !== 'undefined' && x instanceof HTMLElement) {
        return true;
    ***REMOVED***
    return typeof x.nodeName === 'string' && typeof x.getAttribute === 'function';
***REMOVED***

function inspectString(str, opts) {
    if (str.length > opts.maxStringLength) {
        var remaining = str.length - opts.maxStringLength;
        var trailer = '... ' + remaining + ' more character' + (remaining > 1 ? 's' : '');
        return inspectString($slice.call(str, 0, opts.maxStringLength), opts) + trailer;
    ***REMOVED***
    var quoteRE = quoteREs[opts.quoteStyle || 'single'];
    quoteRE.lastIndex = 0;
    // eslint-disable-next-line no-control-regex
    var s = $replace.call($replace.call(str, quoteRE, '\\$1'), /[\x00-\x1f]/g, lowbyte);
    return wrapQuotes(s, 'single', opts);
***REMOVED***

function lowbyte(c) {
    var n = c.charCodeAt(0);
    var x = {
        8: 'b',
        9: 't',
        10: 'n',
        12: 'f',
        13: 'r'
    ***REMOVED***[n];
    if (x) { return '\\' + x; ***REMOVED***
    return '\\x' + (n < 0x10 ? '0' : '') + $toUpperCase.call(n.toString(16));
***REMOVED***

function markBoxed(str) {
    return 'Object(' + str + ')';
***REMOVED***

function weakCollectionOf(type) {
    return type + ' { ? ***REMOVED***';
***REMOVED***

function collectionOf(type, size, entries, indent) {
    var joinedEntries = indent ? indentedJoin(entries, indent) : $join.call(entries, ', ');
    return type + ' (' + size + ') {' + joinedEntries + '***REMOVED***';
***REMOVED***

function singleLineValues(xs) {
    for (var i = 0; i < xs.length; i++) {
        if (indexOf(xs[i], '\n') >= 0) {
            return false;
        ***REMOVED***
    ***REMOVED***
    return true;
***REMOVED***

function getIndent(opts, depth) {
    var baseIndent;
    if (opts.indent === '\t') {
        baseIndent = '\t';
    ***REMOVED*** else if (typeof opts.indent === 'number' && opts.indent > 0) {
        baseIndent = $join.call(Array(opts.indent + 1), ' ');
    ***REMOVED*** else {
        return null;
    ***REMOVED***
    return {
        base: baseIndent,
        prev: $join.call(Array(depth + 1), baseIndent)
    ***REMOVED***;
***REMOVED***

function indentedJoin(xs, indent) {
    if (xs.length === 0) { return ''; ***REMOVED***
    var lineJoiner = '\n' + indent.prev + indent.base;
    return lineJoiner + $join.call(xs, ',' + lineJoiner) + '\n' + indent.prev;
***REMOVED***

function arrObjKeys(obj, inspect) {
    var isArr = isArray(obj);
    var xs = [];
    if (isArr) {
        xs.length = obj.length;
        for (var i = 0; i < obj.length; i++) {
            xs[i] = has(obj, i) ? inspect(obj[i], obj) : '';
        ***REMOVED***
    ***REMOVED***
    var syms = typeof gOPS === 'function' ? gOPS(obj) : [];
    var symMap;
    if (hasShammedSymbols) {
        symMap = {***REMOVED***;
        for (var k = 0; k < syms.length; k++) {
            symMap['$' + syms[k]] = syms[k];
        ***REMOVED***
    ***REMOVED***

    for (var key in obj) { // eslint-disable-line no-restricted-syntax
        if (!has(obj, key)) { continue; ***REMOVED*** // eslint-disable-line no-restricted-syntax, no-continue
        if (isArr && String(Number(key)) === key && key < obj.length) { continue; ***REMOVED*** // eslint-disable-line no-restricted-syntax, no-continue
        if (hasShammedSymbols && symMap['$' + key] instanceof Symbol) {
            // this is to prevent shammed Symbols, which are stored as strings, from being included in the string key section
            continue; // eslint-disable-line no-restricted-syntax, no-continue
        ***REMOVED*** else if ($test.call(/[^\w$]/, key)) {
            xs.push(inspect(key, obj) + ': ' + inspect(obj[key], obj));
        ***REMOVED*** else {
            xs.push(key + ': ' + inspect(obj[key], obj));
        ***REMOVED***
    ***REMOVED***
    if (typeof gOPS === 'function') {
        for (var j = 0; j < syms.length; j++) {
            if (isEnumerable.call(obj, syms[j])) {
                xs.push('[' + inspect(syms[j]) + ']: ' + inspect(obj[syms[j]], obj));
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    return xs;
***REMOVED***
