// @babel/types/lib/asserts/generated/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.assertAccessor = assertAccessor;
exports.assertAnyTypeAnnotation = assertAnyTypeAnnotation;
exports.assertArgumentPlaceholder = assertArgumentPlaceholder;
exports.assertArrayExpression = assertArrayExpression;
exports.assertArrayPattern = assertArrayPattern;
exports.assertArrayTypeAnnotation = assertArrayTypeAnnotation;
exports.assertArrowFunctionExpression = assertArrowFunctionExpression;
exports.assertAssignmentExpression = assertAssignmentExpression;
exports.assertAssignmentPattern = assertAssignmentPattern;
exports.assertAwaitExpression = assertAwaitExpression;
exports.assertBigIntLiteral = assertBigIntLiteral;
exports.assertBinary = assertBinary;
exports.assertBinaryExpression = assertBinaryExpression;
exports.assertBindExpression = assertBindExpression;
exports.assertBlock = assertBlock;
exports.assertBlockParent = assertBlockParent;
exports.assertBlockStatement = assertBlockStatement;
exports.assertBooleanLiteral = assertBooleanLiteral;
exports.assertBooleanLiteralTypeAnnotation = assertBooleanLiteralTypeAnnotation;
exports.assertBooleanTypeAnnotation = assertBooleanTypeAnnotation;
exports.assertBreakStatement = assertBreakStatement;
exports.assertCallExpression = assertCallExpression;
exports.assertCatchClause = assertCatchClause;
exports.assertClass = assertClass;
exports.assertClassAccessorProperty = assertClassAccessorProperty;
exports.assertClassBody = assertClassBody;
exports.assertClassDeclaration = assertClassDeclaration;
exports.assertClassExpression = assertClassExpression;
exports.assertClassImplements = assertClassImplements;
exports.assertClassMethod = assertClassMethod;
exports.assertClassPrivateMethod = assertClassPrivateMethod;
exports.assertClassPrivateProperty = assertClassPrivateProperty;
exports.assertClassProperty = assertClassProperty;
exports.assertCompletionStatement = assertCompletionStatement;
exports.assertConditional = assertConditional;
exports.assertConditionalExpression = assertConditionalExpression;
exports.assertContinueStatement = assertContinueStatement;
exports.assertDebuggerStatement = assertDebuggerStatement;
exports.assertDecimalLiteral = assertDecimalLiteral;
exports.assertDeclaration = assertDeclaration;
exports.assertDeclareClass = assertDeclareClass;
exports.assertDeclareExportAllDeclaration = assertDeclareExportAllDeclaration;
exports.assertDeclareExportDeclaration = assertDeclareExportDeclaration;
exports.assertDeclareFunction = assertDeclareFunction;
exports.assertDeclareInterface = assertDeclareInterface;
exports.assertDeclareModule = assertDeclareModule;
exports.assertDeclareModuleExports = assertDeclareModuleExports;
exports.assertDeclareOpaqueType = assertDeclareOpaqueType;
exports.assertDeclareTypeAlias = assertDeclareTypeAlias;
exports.assertDeclareVariable = assertDeclareVariable;
exports.assertDeclaredPredicate = assertDeclaredPredicate;
exports.assertDecorator = assertDecorator;
exports.assertDirective = assertDirective;
exports.assertDirectiveLiteral = assertDirectiveLiteral;
exports.assertDoExpression = assertDoExpression;
exports.assertDoWhileStatement = assertDoWhileStatement;
exports.assertEmptyStatement = assertEmptyStatement;
exports.assertEmptyTypeAnnotation = assertEmptyTypeAnnotation;
exports.assertEnumBody = assertEnumBody;
exports.assertEnumBooleanBody = assertEnumBooleanBody;
exports.assertEnumBooleanMember = assertEnumBooleanMember;
exports.assertEnumDeclaration = assertEnumDeclaration;
exports.assertEnumDefaultedMember = assertEnumDefaultedMember;
exports.assertEnumMember = assertEnumMember;
exports.assertEnumNumberBody = assertEnumNumberBody;
exports.assertEnumNumberMember = assertEnumNumberMember;
exports.assertEnumStringBody = assertEnumStringBody;
exports.assertEnumStringMember = assertEnumStringMember;
exports.assertEnumSymbolBody = assertEnumSymbolBody;
exports.assertExistsTypeAnnotation = assertExistsTypeAnnotation;
exports.assertExportAllDeclaration = assertExportAllDeclaration;
exports.assertExportDeclaration = assertExportDeclaration;
exports.assertExportDefaultDeclaration = assertExportDefaultDeclaration;
exports.assertExportDefaultSpecifier = assertExportDefaultSpecifier;
exports.assertExportNamedDeclaration = assertExportNamedDeclaration;
exports.assertExportNamespaceSpecifier = assertExportNamespaceSpecifier;
exports.assertExportSpecifier = assertExportSpecifier;
exports.assertExpression = assertExpression;
exports.assertExpressionStatement = assertExpressionStatement;
exports.assertExpressionWrapper = assertExpressionWrapper;
exports.assertFile = assertFile;
exports.assertFlow = assertFlow;
exports.assertFlowBaseAnnotation = assertFlowBaseAnnotation;
exports.assertFlowDeclaration = assertFlowDeclaration;
exports.assertFlowPredicate = assertFlowPredicate;
exports.assertFlowType = assertFlowType;
exports.assertFor = assertFor;
exports.assertForInStatement = assertForInStatement;
exports.assertForOfStatement = assertForOfStatement;
exports.assertForStatement = assertForStatement;
exports.assertForXStatement = assertForXStatement;
exports.assertFunction = assertFunction;
exports.assertFunctionDeclaration = assertFunctionDeclaration;
exports.assertFunctionExpression = assertFunctionExpression;
exports.assertFunctionParent = assertFunctionParent;
exports.assertFunctionTypeAnnotation = assertFunctionTypeAnnotation;
exports.assertFunctionTypeParam = assertFunctionTypeParam;
exports.assertGenericTypeAnnotation = assertGenericTypeAnnotation;
exports.assertIdentifier = assertIdentifier;
exports.assertIfStatement = assertIfStatement;
exports.assertImmutable = assertImmutable;
exports.assertImport = assertImport;
exports.assertImportAttribute = assertImportAttribute;
exports.assertImportDeclaration = assertImportDeclaration;
exports.assertImportDefaultSpecifier = assertImportDefaultSpecifier;
exports.assertImportExpression = assertImportExpression;
exports.assertImportNamespaceSpecifier = assertImportNamespaceSpecifier;
exports.assertImportOrExportDeclaration = assertImportOrExportDeclaration;
exports.assertImportSpecifier = assertImportSpecifier;
exports.assertIndexedAccessType = assertIndexedAccessType;
exports.assertInferredPredicate = assertInferredPredicate;
exports.assertInterfaceDeclaration = assertInterfaceDeclaration;
exports.assertInterfaceExtends = assertInterfaceExtends;
exports.assertInterfaceTypeAnnotation = assertInterfaceTypeAnnotation;
exports.assertInterpreterDirective = assertInterpreterDirective;
exports.assertIntersectionTypeAnnotation = assertIntersectionTypeAnnotation;
exports.assertJSX = assertJSX;
exports.assertJSXAttribute = assertJSXAttribute;
exports.assertJSXClosingElement = assertJSXClosingElement;
exports.assertJSXClosingFragment = assertJSXClosingFragment;
exports.assertJSXElement = assertJSXElement;
exports.assertJSXEmptyExpression = assertJSXEmptyExpression;
exports.assertJSXExpressionContainer = assertJSXExpressionContainer;
exports.assertJSXFragment = assertJSXFragment;
exports.assertJSXIdentifier = assertJSXIdentifier;
exports.assertJSXMemberExpression = assertJSXMemberExpression;
exports.assertJSXNamespacedName = assertJSXNamespacedName;
exports.assertJSXOpeningElement = assertJSXOpeningElement;
exports.assertJSXOpeningFragment = assertJSXOpeningFragment;
exports.assertJSXSpreadAttribute = assertJSXSpreadAttribute;
exports.assertJSXSpreadChild = assertJSXSpreadChild;
exports.assertJSXText = assertJSXText;
exports.assertLVal = assertLVal;
exports.assertLabeledStatement = assertLabeledStatement;
exports.assertLiteral = assertLiteral;
exports.assertLogicalExpression = assertLogicalExpression;
exports.assertLoop = assertLoop;
exports.assertMemberExpression = assertMemberExpression;
exports.assertMetaProperty = assertMetaProperty;
exports.assertMethod = assertMethod;
exports.assertMiscellaneous = assertMiscellaneous;
exports.assertMixedTypeAnnotation = assertMixedTypeAnnotation;
exports.assertModuleDeclaration = assertModuleDeclaration;
exports.assertModuleExpression = assertModuleExpression;
exports.assertModuleSpecifier = assertModuleSpecifier;
exports.assertNewExpression = assertNewExpression;
exports.assertNoop = assertNoop;
exports.assertNullLiteral = assertNullLiteral;
exports.assertNullLiteralTypeAnnotation = assertNullLiteralTypeAnnotation;
exports.assertNullableTypeAnnotation = assertNullableTypeAnnotation;
exports.assertNumberLiteral = assertNumberLiteral;
exports.assertNumberLiteralTypeAnnotation = assertNumberLiteralTypeAnnotation;
exports.assertNumberTypeAnnotation = assertNumberTypeAnnotation;
exports.assertNumericLiteral = assertNumericLiteral;
exports.assertObjectExpression = assertObjectExpression;
exports.assertObjectMember = assertObjectMember;
exports.assertObjectMethod = assertObjectMethod;
exports.assertObjectPattern = assertObjectPattern;
exports.assertObjectProperty = assertObjectProperty;
exports.assertObjectTypeAnnotation = assertObjectTypeAnnotation;
exports.assertObjectTypeCallProperty = assertObjectTypeCallProperty;
exports.assertObjectTypeIndexer = assertObjectTypeIndexer;
exports.assertObjectTypeInternalSlot = assertObjectTypeInternalSlot;
exports.assertObjectTypeProperty = assertObjectTypeProperty;
exports.assertObjectTypeSpreadProperty = assertObjectTypeSpreadProperty;
exports.assertOpaqueType = assertOpaqueType;
exports.assertOptionalCallExpression = assertOptionalCallExpression;
exports.assertOptionalIndexedAccessType = assertOptionalIndexedAccessType;
exports.assertOptionalMemberExpression = assertOptionalMemberExpression;
exports.assertParenthesizedExpression = assertParenthesizedExpression;
exports.assertPattern = assertPattern;
exports.assertPatternLike = assertPatternLike;
exports.assertPipelineBareFunction = assertPipelineBareFunction;
exports.assertPipelinePrimaryTopicReference = assertPipelinePrimaryTopicReference;
exports.assertPipelineTopicExpression = assertPipelineTopicExpression;
exports.assertPlaceholder = assertPlaceholder;
exports.assertPrivate = assertPrivate;
exports.assertPrivateName = assertPrivateName;
exports.assertProgram = assertProgram;
exports.assertProperty = assertProperty;
exports.assertPureish = assertPureish;
exports.assertQualifiedTypeIdentifier = assertQualifiedTypeIdentifier;
exports.assertRecordExpression = assertRecordExpression;
exports.assertRegExpLiteral = assertRegExpLiteral;
exports.assertRegexLiteral = assertRegexLiteral;
exports.assertRestElement = assertRestElement;
exports.assertRestProperty = assertRestProperty;
exports.assertReturnStatement = assertReturnStatement;
exports.assertScopable = assertScopable;
exports.assertSequenceExpression = assertSequenceExpression;
exports.assertSpreadElement = assertSpreadElement;
exports.assertSpreadProperty = assertSpreadProperty;
exports.assertStandardized = assertStandardized;
exports.assertStatement = assertStatement;
exports.assertStaticBlock = assertStaticBlock;
exports.assertStringLiteral = assertStringLiteral;
exports.assertStringLiteralTypeAnnotation = assertStringLiteralTypeAnnotation;
exports.assertStringTypeAnnotation = assertStringTypeAnnotation;
exports.assertSuper = assertSuper;
exports.assertSwitchCase = assertSwitchCase;
exports.assertSwitchStatement = assertSwitchStatement;
exports.assertSymbolTypeAnnotation = assertSymbolTypeAnnotation;
exports.assertTSAnyKeyword = assertTSAnyKeyword;
exports.assertTSArrayType = assertTSArrayType;
exports.assertTSAsExpression = assertTSAsExpression;
exports.assertTSBaseType = assertTSBaseType;
exports.assertTSBigIntKeyword = assertTSBigIntKeyword;
exports.assertTSBooleanKeyword = assertTSBooleanKeyword;
exports.assertTSCallSignatureDeclaration = assertTSCallSignatureDeclaration;
exports.assertTSConditionalType = assertTSConditionalType;
exports.assertTSConstructSignatureDeclaration = assertTSConstructSignatureDeclaration;
exports.assertTSConstructorType = assertTSConstructorType;
exports.assertTSDeclareFunction = assertTSDeclareFunction;
exports.assertTSDeclareMethod = assertTSDeclareMethod;
exports.assertTSEntityName = assertTSEntityName;
exports.assertTSEnumBody = assertTSEnumBody;
exports.assertTSEnumDeclaration = assertTSEnumDeclaration;
exports.assertTSEnumMember = assertTSEnumMember;
exports.assertTSExportAssignment = assertTSExportAssignment;
exports.assertTSExpressionWithTypeArguments = assertTSExpressionWithTypeArguments;
exports.assertTSExternalModuleReference = assertTSExternalModuleReference;
exports.assertTSFunctionType = assertTSFunctionType;
exports.assertTSImportEqualsDeclaration = assertTSImportEqualsDeclaration;
exports.assertTSImportType = assertTSImportType;
exports.assertTSIndexSignature = assertTSIndexSignature;
exports.assertTSIndexedAccessType = assertTSIndexedAccessType;
exports.assertTSInferType = assertTSInferType;
exports.assertTSInstantiationExpression = assertTSInstantiationExpression;
exports.assertTSInterfaceBody = assertTSInterfaceBody;
exports.assertTSInterfaceDeclaration = assertTSInterfaceDeclaration;
exports.assertTSIntersectionType = assertTSIntersectionType;
exports.assertTSIntrinsicKeyword = assertTSIntrinsicKeyword;
exports.assertTSLiteralType = assertTSLiteralType;
exports.assertTSMappedType = assertTSMappedType;
exports.assertTSMethodSignature = assertTSMethodSignature;
exports.assertTSModuleBlock = assertTSModuleBlock;
exports.assertTSModuleDeclaration = assertTSModuleDeclaration;
exports.assertTSNamedTupleMember = assertTSNamedTupleMember;
exports.assertTSNamespaceExportDeclaration = assertTSNamespaceExportDeclaration;
exports.assertTSNeverKeyword = assertTSNeverKeyword;
exports.assertTSNonNullExpression = assertTSNonNullExpression;
exports.assertTSNullKeyword = assertTSNullKeyword;
exports.assertTSNumberKeyword = assertTSNumberKeyword;
exports.assertTSObjectKeyword = assertTSObjectKeyword;
exports.assertTSOptionalType = assertTSOptionalType;
exports.assertTSParameterProperty = assertTSParameterProperty;
exports.assertTSParenthesizedType = assertTSParenthesizedType;
exports.assertTSPropertySignature = assertTSPropertySignature;
exports.assertTSQualifiedName = assertTSQualifiedName;
exports.assertTSRestType = assertTSRestType;
exports.assertTSSatisfiesExpression = assertTSSatisfiesExpression;
exports.assertTSStringKeyword = assertTSStringKeyword;
exports.assertTSSymbolKeyword = assertTSSymbolKeyword;
exports.assertTSTemplateLiteralType = assertTSTemplateLiteralType;
exports.assertTSThisType = assertTSThisType;
exports.assertTSTupleType = assertTSTupleType;
exports.assertTSType = assertTSType;
exports.assertTSTypeAliasDeclaration = assertTSTypeAliasDeclaration;
exports.assertTSTypeAnnotation = assertTSTypeAnnotation;
exports.assertTSTypeAssertion = assertTSTypeAssertion;
exports.assertTSTypeElement = assertTSTypeElement;
exports.assertTSTypeLiteral = assertTSTypeLiteral;
exports.assertTSTypeOperator = assertTSTypeOperator;
exports.assertTSTypeParameter = assertTSTypeParameter;
exports.assertTSTypeParameterDeclaration = assertTSTypeParameterDeclaration;
exports.assertTSTypeParameterInstantiation = assertTSTypeParameterInstantiation;
exports.assertTSTypePredicate = assertTSTypePredicate;
exports.assertTSTypeQuery = assertTSTypeQuery;
exports.assertTSTypeReference = assertTSTypeReference;
exports.assertTSUndefinedKeyword = assertTSUndefinedKeyword;
exports.assertTSUnionType = assertTSUnionType;
exports.assertTSUnknownKeyword = assertTSUnknownKeyword;
exports.assertTSVoidKeyword = assertTSVoidKeyword;
exports.assertTaggedTemplateExpression = assertTaggedTemplateExpression;
exports.assertTemplateElement = assertTemplateElement;
exports.assertTemplateLiteral = assertTemplateLiteral;
exports.assertTerminatorless = assertTerminatorless;
exports.assertThisExpression = assertThisExpression;
exports.assertThisTypeAnnotation = assertThisTypeAnnotation;
exports.assertThrowStatement = assertThrowStatement;
exports.assertTopicReference = assertTopicReference;
exports.assertTryStatement = assertTryStatement;
exports.assertTupleExpression = assertTupleExpression;
exports.assertTupleTypeAnnotation = assertTupleTypeAnnotation;
exports.assertTypeAlias = assertTypeAlias;
exports.assertTypeAnnotation = assertTypeAnnotation;
exports.assertTypeCastExpression = assertTypeCastExpression;
exports.assertTypeParameter = assertTypeParameter;
exports.assertTypeParameterDeclaration = assertTypeParameterDeclaration;
exports.assertTypeParameterInstantiation = assertTypeParameterInstantiation;
exports.assertTypeScript = assertTypeScript;
exports.assertTypeofTypeAnnotation = assertTypeofTypeAnnotation;
exports.assertUnaryExpression = assertUnaryExpression;
exports.assertUnaryLike = assertUnaryLike;
exports.assertUnionTypeAnnotation = assertUnionTypeAnnotation;
exports.assertUpdateExpression = assertUpdateExpression;
exports.assertUserWhitespacable = assertUserWhitespacable;
exports.assertV8IntrinsicIdentifier = assertV8IntrinsicIdentifier;
exports.assertVariableDeclaration = assertVariableDeclaration;
exports.assertVariableDeclarator = assertVariableDeclarator;
exports.assertVariance = assertVariance;
exports.assertVoidTypeAnnotation = assertVoidTypeAnnotation;
exports.assertWhile = assertWhile;
exports.assertWhileStatement = assertWhileStatement;
exports.assertWithStatement = assertWithStatement;
exports.assertYieldExpression = assertYieldExpression;
var _is = require("../../validators/is.js");
var _deprecationWarning = require("../../utils/deprecationWarning.js");
function assert(type, node, opts) {
  if (!(0, _is.default)(type, node, opts)) {
    throw new Error(`Expected type "${type***REMOVED***" with option ${JSON.stringify(opts)***REMOVED***, ` + `but instead got "${node.type***REMOVED***".`);
  ***REMOVED***
***REMOVED***
function assertArrayExpression(node, opts) {
  assert("ArrayExpression", node, opts);
***REMOVED***
function assertAssignmentExpression(node, opts) {
  assert("AssignmentExpression", node, opts);
***REMOVED***
function assertBinaryExpression(node, opts) {
  assert("BinaryExpression", node, opts);
***REMOVED***
function assertInterpreterDirective(node, opts) {
  assert("InterpreterDirective", node, opts);
***REMOVED***
function assertDirective(node, opts) {
  assert("Directive", node, opts);
***REMOVED***
function assertDirectiveLiteral(node, opts) {
  assert("DirectiveLiteral", node, opts);
***REMOVED***
function assertBlockStatement(node, opts) {
  assert("BlockStatement", node, opts);
***REMOVED***
function assertBreakStatement(node, opts) {
  assert("BreakStatement", node, opts);
***REMOVED***
function assertCallExpression(node, opts) {
  assert("CallExpression", node, opts);
***REMOVED***
function assertCatchClause(node, opts) {
  assert("CatchClause", node, opts);
***REMOVED***
function assertConditionalExpression(node, opts) {
  assert("ConditionalExpression", node, opts);
***REMOVED***
function assertContinueStatement(node, opts) {
  assert("ContinueStatement", node, opts);
***REMOVED***
function assertDebuggerStatement(node, opts) {
  assert("DebuggerStatement", node, opts);
***REMOVED***
function assertDoWhileStatement(node, opts) {
  assert("DoWhileStatement", node, opts);
***REMOVED***
function assertEmptyStatement(node, opts) {
  assert("EmptyStatement", node, opts);
***REMOVED***
function assertExpressionStatement(node, opts) {
  assert("ExpressionStatement", node, opts);
***REMOVED***
function assertFile(node, opts) {
  assert("File", node, opts);
***REMOVED***
function assertForInStatement(node, opts) {
  assert("ForInStatement", node, opts);
***REMOVED***
function assertForStatement(node, opts) {
  assert("ForStatement", node, opts);
***REMOVED***
function assertFunctionDeclaration(node, opts) {
  assert("FunctionDeclaration", node, opts);
***REMOVED***
function assertFunctionExpression(node, opts) {
  assert("FunctionExpression", node, opts);
***REMOVED***
function assertIdentifier(node, opts) {
  assert("Identifier", node, opts);
***REMOVED***
function assertIfStatement(node, opts) {
  assert("IfStatement", node, opts);
***REMOVED***
function assertLabeledStatement(node, opts) {
  assert("LabeledStatement", node, opts);
***REMOVED***
function assertStringLiteral(node, opts) {
  assert("StringLiteral", node, opts);
***REMOVED***
function assertNumericLiteral(node, opts) {
  assert("NumericLiteral", node, opts);
***REMOVED***
function assertNullLiteral(node, opts) {
  assert("NullLiteral", node, opts);
***REMOVED***
function assertBooleanLiteral(node, opts) {
  assert("BooleanLiteral", node, opts);
***REMOVED***
function assertRegExpLiteral(node, opts) {
  assert("RegExpLiteral", node, opts);
***REMOVED***
function assertLogicalExpression(node, opts) {
  assert("LogicalExpression", node, opts);
***REMOVED***
function assertMemberExpression(node, opts) {
  assert("MemberExpression", node, opts);
***REMOVED***
function assertNewExpression(node, opts) {
  assert("NewExpression", node, opts);
***REMOVED***
function assertProgram(node, opts) {
  assert("Program", node, opts);
***REMOVED***
function assertObjectExpression(node, opts) {
  assert("ObjectExpression", node, opts);
***REMOVED***
function assertObjectMethod(node, opts) {
  assert("ObjectMethod", node, opts);
***REMOVED***
function assertObjectProperty(node, opts) {
  assert("ObjectProperty", node, opts);
***REMOVED***
function assertRestElement(node, opts) {
  assert("RestElement", node, opts);
***REMOVED***
function assertReturnStatement(node, opts) {
  assert("ReturnStatement", node, opts);
***REMOVED***
function assertSequenceExpression(node, opts) {
  assert("SequenceExpression", node, opts);
***REMOVED***
function assertParenthesizedExpression(node, opts) {
  assert("ParenthesizedExpression", node, opts);
***REMOVED***
function assertSwitchCase(node, opts) {
  assert("SwitchCase", node, opts);
***REMOVED***
function assertSwitchStatement(node, opts) {
  assert("SwitchStatement", node, opts);
***REMOVED***
function assertThisExpression(node, opts) {
  assert("ThisExpression", node, opts);
***REMOVED***
function assertThrowStatement(node, opts) {
  assert("ThrowStatement", node, opts);
***REMOVED***
function assertTryStatement(node, opts) {
  assert("TryStatement", node, opts);
***REMOVED***
function assertUnaryExpression(node, opts) {
  assert("UnaryExpression", node, opts);
***REMOVED***
function assertUpdateExpression(node, opts) {
  assert("UpdateExpression", node, opts);
***REMOVED***
function assertVariableDeclaration(node, opts) {
  assert("VariableDeclaration", node, opts);
***REMOVED***
function assertVariableDeclarator(node, opts) {
  assert("VariableDeclarator", node, opts);
***REMOVED***
function assertWhileStatement(node, opts) {
  assert("WhileStatement", node, opts);
***REMOVED***
function assertWithStatement(node, opts) {
  assert("WithStatement", node, opts);
***REMOVED***
function assertAssignmentPattern(node, opts) {
  assert("AssignmentPattern", node, opts);
***REMOVED***
function assertArrayPattern(node, opts) {
  assert("ArrayPattern", node, opts);
***REMOVED***
function assertArrowFunctionExpression(node, opts) {
  assert("ArrowFunctionExpression", node, opts);
***REMOVED***
function assertClassBody(node, opts) {
  assert("ClassBody", node, opts);
***REMOVED***
function assertClassExpression(node, opts) {
  assert("ClassExpression", node, opts);
***REMOVED***
function assertClassDeclaration(node, opts) {
  assert("ClassDeclaration", node, opts);
***REMOVED***
function assertExportAllDeclaration(node, opts) {
  assert("ExportAllDeclaration", node, opts);
***REMOVED***
function assertExportDefaultDeclaration(node, opts) {
  assert("ExportDefaultDeclaration", node, opts);
***REMOVED***
function assertExportNamedDeclaration(node, opts) {
  assert("ExportNamedDeclaration", node, opts);
***REMOVED***
function assertExportSpecifier(node, opts) {
  assert("ExportSpecifier", node, opts);
***REMOVED***
function assertForOfStatement(node, opts) {
  assert("ForOfStatement", node, opts);
***REMOVED***
function assertImportDeclaration(node, opts) {
  assert("ImportDeclaration", node, opts);
***REMOVED***
function assertImportDefaultSpecifier(node, opts) {
  assert("ImportDefaultSpecifier", node, opts);
***REMOVED***
function assertImportNamespaceSpecifier(node, opts) {
  assert("ImportNamespaceSpecifier", node, opts);
***REMOVED***
function assertImportSpecifier(node, opts) {
  assert("ImportSpecifier", node, opts);
***REMOVED***
function assertImportExpression(node, opts) {
  assert("ImportExpression", node, opts);
***REMOVED***
function assertMetaProperty(node, opts) {
  assert("MetaProperty", node, opts);
***REMOVED***
function assertClassMethod(node, opts) {
  assert("ClassMethod", node, opts);
***REMOVED***
function assertObjectPattern(node, opts) {
  assert("ObjectPattern", node, opts);
***REMOVED***
function assertSpreadElement(node, opts) {
  assert("SpreadElement", node, opts);
***REMOVED***
function assertSuper(node, opts) {
  assert("Super", node, opts);
***REMOVED***
function assertTaggedTemplateExpression(node, opts) {
  assert("TaggedTemplateExpression", node, opts);
***REMOVED***
function assertTemplateElement(node, opts) {
  assert("TemplateElement", node, opts);
***REMOVED***
function assertTemplateLiteral(node, opts) {
  assert("TemplateLiteral", node, opts);
***REMOVED***
function assertYieldExpression(node, opts) {
  assert("YieldExpression", node, opts);
***REMOVED***
function assertAwaitExpression(node, opts) {
  assert("AwaitExpression", node, opts);
***REMOVED***
function assertImport(node, opts) {
  assert("Import", node, opts);
***REMOVED***
function assertBigIntLiteral(node, opts) {
  assert("BigIntLiteral", node, opts);
***REMOVED***
function assertExportNamespaceSpecifier(node, opts) {
  assert("ExportNamespaceSpecifier", node, opts);
***REMOVED***
function assertOptionalMemberExpression(node, opts) {
  assert("OptionalMemberExpression", node, opts);
***REMOVED***
function assertOptionalCallExpression(node, opts) {
  assert("OptionalCallExpression", node, opts);
***REMOVED***
function assertClassProperty(node, opts) {
  assert("ClassProperty", node, opts);
***REMOVED***
function assertClassAccessorProperty(node, opts) {
  assert("ClassAccessorProperty", node, opts);
***REMOVED***
function assertClassPrivateProperty(node, opts) {
  assert("ClassPrivateProperty", node, opts);
***REMOVED***
function assertClassPrivateMethod(node, opts) {
  assert("ClassPrivateMethod", node, opts);
***REMOVED***
function assertPrivateName(node, opts) {
  assert("PrivateName", node, opts);
***REMOVED***
function assertStaticBlock(node, opts) {
  assert("StaticBlock", node, opts);
***REMOVED***
function assertImportAttribute(node, opts) {
  assert("ImportAttribute", node, opts);
***REMOVED***
function assertAnyTypeAnnotation(node, opts) {
  assert("AnyTypeAnnotation", node, opts);
***REMOVED***
function assertArrayTypeAnnotation(node, opts) {
  assert("ArrayTypeAnnotation", node, opts);
***REMOVED***
function assertBooleanTypeAnnotation(node, opts) {
  assert("BooleanTypeAnnotation", node, opts);
***REMOVED***
function assertBooleanLiteralTypeAnnotation(node, opts) {
  assert("BooleanLiteralTypeAnnotation", node, opts);
***REMOVED***
function assertNullLiteralTypeAnnotation(node, opts) {
  assert("NullLiteralTypeAnnotation", node, opts);
***REMOVED***
function assertClassImplements(node, opts) {
  assert("ClassImplements", node, opts);
***REMOVED***
function assertDeclareClass(node, opts) {
  assert("DeclareClass", node, opts);
***REMOVED***
function assertDeclareFunction(node, opts) {
  assert("DeclareFunction", node, opts);
***REMOVED***
function assertDeclareInterface(node, opts) {
  assert("DeclareInterface", node, opts);
***REMOVED***
function assertDeclareModule(node, opts) {
  assert("DeclareModule", node, opts);
***REMOVED***
function assertDeclareModuleExports(node, opts) {
  assert("DeclareModuleExports", node, opts);
***REMOVED***
function assertDeclareTypeAlias(node, opts) {
  assert("DeclareTypeAlias", node, opts);
***REMOVED***
function assertDeclareOpaqueType(node, opts) {
  assert("DeclareOpaqueType", node, opts);
***REMOVED***
function assertDeclareVariable(node, opts) {
  assert("DeclareVariable", node, opts);
***REMOVED***
function assertDeclareExportDeclaration(node, opts) {
  assert("DeclareExportDeclaration", node, opts);
***REMOVED***
function assertDeclareExportAllDeclaration(node, opts) {
  assert("DeclareExportAllDeclaration", node, opts);
***REMOVED***
function assertDeclaredPredicate(node, opts) {
  assert("DeclaredPredicate", node, opts);
***REMOVED***
function assertExistsTypeAnnotation(node, opts) {
  assert("ExistsTypeAnnotation", node, opts);
***REMOVED***
function assertFunctionTypeAnnotation(node, opts) {
  assert("FunctionTypeAnnotation", node, opts);
***REMOVED***
function assertFunctionTypeParam(node, opts) {
  assert("FunctionTypeParam", node, opts);
***REMOVED***
function assertGenericTypeAnnotation(node, opts) {
  assert("GenericTypeAnnotation", node, opts);
***REMOVED***
function assertInferredPredicate(node, opts) {
  assert("InferredPredicate", node, opts);
***REMOVED***
function assertInterfaceExtends(node, opts) {
  assert("InterfaceExtends", node, opts);
***REMOVED***
function assertInterfaceDeclaration(node, opts) {
  assert("InterfaceDeclaration", node, opts);
***REMOVED***
function assertInterfaceTypeAnnotation(node, opts) {
  assert("InterfaceTypeAnnotation", node, opts);
***REMOVED***
function assertIntersectionTypeAnnotation(node, opts) {
  assert("IntersectionTypeAnnotation", node, opts);
***REMOVED***
function assertMixedTypeAnnotation(node, opts) {
  assert("MixedTypeAnnotation", node, opts);
***REMOVED***
function assertEmptyTypeAnnotation(node, opts) {
  assert("EmptyTypeAnnotation", node, opts);
***REMOVED***
function assertNullableTypeAnnotation(node, opts) {
  assert("NullableTypeAnnotation", node, opts);
***REMOVED***
function assertNumberLiteralTypeAnnotation(node, opts) {
  assert("NumberLiteralTypeAnnotation", node, opts);
***REMOVED***
function assertNumberTypeAnnotation(node, opts) {
  assert("NumberTypeAnnotation", node, opts);
***REMOVED***
function assertObjectTypeAnnotation(node, opts) {
  assert("ObjectTypeAnnotation", node, opts);
***REMOVED***
function assertObjectTypeInternalSlot(node, opts) {
  assert("ObjectTypeInternalSlot", node, opts);
***REMOVED***
function assertObjectTypeCallProperty(node, opts) {
  assert("ObjectTypeCallProperty", node, opts);
***REMOVED***
function assertObjectTypeIndexer(node, opts) {
  assert("ObjectTypeIndexer", node, opts);
***REMOVED***
function assertObjectTypeProperty(node, opts) {
  assert("ObjectTypeProperty", node, opts);
***REMOVED***
function assertObjectTypeSpreadProperty(node, opts) {
  assert("ObjectTypeSpreadProperty", node, opts);
***REMOVED***
function assertOpaqueType(node, opts) {
  assert("OpaqueType", node, opts);
***REMOVED***
function assertQualifiedTypeIdentifier(node, opts) {
  assert("QualifiedTypeIdentifier", node, opts);
***REMOVED***
function assertStringLiteralTypeAnnotation(node, opts) {
  assert("StringLiteralTypeAnnotation", node, opts);
***REMOVED***
function assertStringTypeAnnotation(node, opts) {
  assert("StringTypeAnnotation", node, opts);
***REMOVED***
function assertSymbolTypeAnnotation(node, opts) {
  assert("SymbolTypeAnnotation", node, opts);
***REMOVED***
function assertThisTypeAnnotation(node, opts) {
  assert("ThisTypeAnnotation", node, opts);
***REMOVED***
function assertTupleTypeAnnotation(node, opts) {
  assert("TupleTypeAnnotation", node, opts);
***REMOVED***
function assertTypeofTypeAnnotation(node, opts) {
  assert("TypeofTypeAnnotation", node, opts);
***REMOVED***
function assertTypeAlias(node, opts) {
  assert("TypeAlias", node, opts);
***REMOVED***
function assertTypeAnnotation(node, opts) {
  assert("TypeAnnotation", node, opts);
***REMOVED***
function assertTypeCastExpression(node, opts) {
  assert("TypeCastExpression", node, opts);
***REMOVED***
function assertTypeParameter(node, opts) {
  assert("TypeParameter", node, opts);
***REMOVED***
function assertTypeParameterDeclaration(node, opts) {
  assert("TypeParameterDeclaration", node, opts);
***REMOVED***
function assertTypeParameterInstantiation(node, opts) {
  assert("TypeParameterInstantiation", node, opts);
***REMOVED***
function assertUnionTypeAnnotation(node, opts) {
  assert("UnionTypeAnnotation", node, opts);
***REMOVED***
function assertVariance(node, opts) {
  assert("Variance", node, opts);
***REMOVED***
function assertVoidTypeAnnotation(node, opts) {
  assert("VoidTypeAnnotation", node, opts);
***REMOVED***
function assertEnumDeclaration(node, opts) {
  assert("EnumDeclaration", node, opts);
***REMOVED***
function assertEnumBooleanBody(node, opts) {
  assert("EnumBooleanBody", node, opts);
***REMOVED***
function assertEnumNumberBody(node, opts) {
  assert("EnumNumberBody", node, opts);
***REMOVED***
function assertEnumStringBody(node, opts) {
  assert("EnumStringBody", node, opts);
***REMOVED***
function assertEnumSymbolBody(node, opts) {
  assert("EnumSymbolBody", node, opts);
***REMOVED***
function assertEnumBooleanMember(node, opts) {
  assert("EnumBooleanMember", node, opts);
***REMOVED***
function assertEnumNumberMember(node, opts) {
  assert("EnumNumberMember", node, opts);
***REMOVED***
function assertEnumStringMember(node, opts) {
  assert("EnumStringMember", node, opts);
***REMOVED***
function assertEnumDefaultedMember(node, opts) {
  assert("EnumDefaultedMember", node, opts);
***REMOVED***
function assertIndexedAccessType(node, opts) {
  assert("IndexedAccessType", node, opts);
***REMOVED***
function assertOptionalIndexedAccessType(node, opts) {
  assert("OptionalIndexedAccessType", node, opts);
***REMOVED***
function assertJSXAttribute(node, opts) {
  assert("JSXAttribute", node, opts);
***REMOVED***
function assertJSXClosingElement(node, opts) {
  assert("JSXClosingElement", node, opts);
***REMOVED***
function assertJSXElement(node, opts) {
  assert("JSXElement", node, opts);
***REMOVED***
function assertJSXEmptyExpression(node, opts) {
  assert("JSXEmptyExpression", node, opts);
***REMOVED***
function assertJSXExpressionContainer(node, opts) {
  assert("JSXExpressionContainer", node, opts);
***REMOVED***
function assertJSXSpreadChild(node, opts) {
  assert("JSXSpreadChild", node, opts);
***REMOVED***
function assertJSXIdentifier(node, opts) {
  assert("JSXIdentifier", node, opts);
***REMOVED***
function assertJSXMemberExpression(node, opts) {
  assert("JSXMemberExpression", node, opts);
***REMOVED***
function assertJSXNamespacedName(node, opts) {
  assert("JSXNamespacedName", node, opts);
***REMOVED***
function assertJSXOpeningElement(node, opts) {
  assert("JSXOpeningElement", node, opts);
***REMOVED***
function assertJSXSpreadAttribute(node, opts) {
  assert("JSXSpreadAttribute", node, opts);
***REMOVED***
function assertJSXText(node, opts) {
  assert("JSXText", node, opts);
***REMOVED***
function assertJSXFragment(node, opts) {
  assert("JSXFragment", node, opts);
***REMOVED***
function assertJSXOpeningFragment(node, opts) {
  assert("JSXOpeningFragment", node, opts);
***REMOVED***
function assertJSXClosingFragment(node, opts) {
  assert("JSXClosingFragment", node, opts);
***REMOVED***
function assertNoop(node, opts) {
  assert("Noop", node, opts);
***REMOVED***
function assertPlaceholder(node, opts) {
  assert("Placeholder", node, opts);
***REMOVED***
function assertV8IntrinsicIdentifier(node, opts) {
  assert("V8IntrinsicIdentifier", node, opts);
***REMOVED***
function assertArgumentPlaceholder(node, opts) {
  assert("ArgumentPlaceholder", node, opts);
***REMOVED***
function assertBindExpression(node, opts) {
  assert("BindExpression", node, opts);
***REMOVED***
function assertDecorator(node, opts) {
  assert("Decorator", node, opts);
***REMOVED***
function assertDoExpression(node, opts) {
  assert("DoExpression", node, opts);
***REMOVED***
function assertExportDefaultSpecifier(node, opts) {
  assert("ExportDefaultSpecifier", node, opts);
***REMOVED***
function assertRecordExpression(node, opts) {
  assert("RecordExpression", node, opts);
***REMOVED***
function assertTupleExpression(node, opts) {
  assert("TupleExpression", node, opts);
***REMOVED***
function assertDecimalLiteral(node, opts) {
  assert("DecimalLiteral", node, opts);
***REMOVED***
function assertModuleExpression(node, opts) {
  assert("ModuleExpression", node, opts);
***REMOVED***
function assertTopicReference(node, opts) {
  assert("TopicReference", node, opts);
***REMOVED***
function assertPipelineTopicExpression(node, opts) {
  assert("PipelineTopicExpression", node, opts);
***REMOVED***
function assertPipelineBareFunction(node, opts) {
  assert("PipelineBareFunction", node, opts);
***REMOVED***
function assertPipelinePrimaryTopicReference(node, opts) {
  assert("PipelinePrimaryTopicReference", node, opts);
***REMOVED***
function assertTSParameterProperty(node, opts) {
  assert("TSParameterProperty", node, opts);
***REMOVED***
function assertTSDeclareFunction(node, opts) {
  assert("TSDeclareFunction", node, opts);
***REMOVED***
function assertTSDeclareMethod(node, opts) {
  assert("TSDeclareMethod", node, opts);
***REMOVED***
function assertTSQualifiedName(node, opts) {
  assert("TSQualifiedName", node, opts);
***REMOVED***
function assertTSCallSignatureDeclaration(node, opts) {
  assert("TSCallSignatureDeclaration", node, opts);
***REMOVED***
function assertTSConstructSignatureDeclaration(node, opts) {
  assert("TSConstructSignatureDeclaration", node, opts);
***REMOVED***
function assertTSPropertySignature(node, opts) {
  assert("TSPropertySignature", node, opts);
***REMOVED***
function assertTSMethodSignature(node, opts) {
  assert("TSMethodSignature", node, opts);
***REMOVED***
function assertTSIndexSignature(node, opts) {
  assert("TSIndexSignature", node, opts);
***REMOVED***
function assertTSAnyKeyword(node, opts) {
  assert("TSAnyKeyword", node, opts);
***REMOVED***
function assertTSBooleanKeyword(node, opts) {
  assert("TSBooleanKeyword", node, opts);
***REMOVED***
function assertTSBigIntKeyword(node, opts) {
  assert("TSBigIntKeyword", node, opts);
***REMOVED***
function assertTSIntrinsicKeyword(node, opts) {
  assert("TSIntrinsicKeyword", node, opts);
***REMOVED***
function assertTSNeverKeyword(node, opts) {
  assert("TSNeverKeyword", node, opts);
***REMOVED***
function assertTSNullKeyword(node, opts) {
  assert("TSNullKeyword", node, opts);
***REMOVED***
function assertTSNumberKeyword(node, opts) {
  assert("TSNumberKeyword", node, opts);
***REMOVED***
function assertTSObjectKeyword(node, opts) {
  assert("TSObjectKeyword", node, opts);
***REMOVED***
function assertTSStringKeyword(node, opts) {
  assert("TSStringKeyword", node, opts);
***REMOVED***
function assertTSSymbolKeyword(node, opts) {
  assert("TSSymbolKeyword", node, opts);
***REMOVED***
function assertTSUndefinedKeyword(node, opts) {
  assert("TSUndefinedKeyword", node, opts);
***REMOVED***
function assertTSUnknownKeyword(node, opts) {
  assert("TSUnknownKeyword", node, opts);
***REMOVED***
function assertTSVoidKeyword(node, opts) {
  assert("TSVoidKeyword", node, opts);
***REMOVED***
function assertTSThisType(node, opts) {
  assert("TSThisType", node, opts);
***REMOVED***
function assertTSFunctionType(node, opts) {
  assert("TSFunctionType", node, opts);
***REMOVED***
function assertTSConstructorType(node, opts) {
  assert("TSConstructorType", node, opts);
***REMOVED***
function assertTSTypeReference(node, opts) {
  assert("TSTypeReference", node, opts);
***REMOVED***
function assertTSTypePredicate(node, opts) {
  assert("TSTypePredicate", node, opts);
***REMOVED***
function assertTSTypeQuery(node, opts) {
  assert("TSTypeQuery", node, opts);
***REMOVED***
function assertTSTypeLiteral(node, opts) {
  assert("TSTypeLiteral", node, opts);
***REMOVED***
function assertTSArrayType(node, opts) {
  assert("TSArrayType", node, opts);
***REMOVED***
function assertTSTupleType(node, opts) {
  assert("TSTupleType", node, opts);
***REMOVED***
function assertTSOptionalType(node, opts) {
  assert("TSOptionalType", node, opts);
***REMOVED***
function assertTSRestType(node, opts) {
  assert("TSRestType", node, opts);
***REMOVED***
function assertTSNamedTupleMember(node, opts) {
  assert("TSNamedTupleMember", node, opts);
***REMOVED***
function assertTSUnionType(node, opts) {
  assert("TSUnionType", node, opts);
***REMOVED***
function assertTSIntersectionType(node, opts) {
  assert("TSIntersectionType", node, opts);
***REMOVED***
function assertTSConditionalType(node, opts) {
  assert("TSConditionalType", node, opts);
***REMOVED***
function assertTSInferType(node, opts) {
  assert("TSInferType", node, opts);
***REMOVED***
function assertTSParenthesizedType(node, opts) {
  assert("TSParenthesizedType", node, opts);
***REMOVED***
function assertTSTypeOperator(node, opts) {
  assert("TSTypeOperator", node, opts);
***REMOVED***
function assertTSIndexedAccessType(node, opts) {
  assert("TSIndexedAccessType", node, opts);
***REMOVED***
function assertTSMappedType(node, opts) {
  assert("TSMappedType", node, opts);
***REMOVED***
function assertTSTemplateLiteralType(node, opts) {
  assert("TSTemplateLiteralType", node, opts);
***REMOVED***
function assertTSLiteralType(node, opts) {
  assert("TSLiteralType", node, opts);
***REMOVED***
function assertTSExpressionWithTypeArguments(node, opts) {
  assert("TSExpressionWithTypeArguments", node, opts);
***REMOVED***
function assertTSInterfaceDeclaration(node, opts) {
  assert("TSInterfaceDeclaration", node, opts);
***REMOVED***
function assertTSInterfaceBody(node, opts) {
  assert("TSInterfaceBody", node, opts);
***REMOVED***
function assertTSTypeAliasDeclaration(node, opts) {
  assert("TSTypeAliasDeclaration", node, opts);
***REMOVED***
function assertTSInstantiationExpression(node, opts) {
  assert("TSInstantiationExpression", node, opts);
***REMOVED***
function assertTSAsExpression(node, opts) {
  assert("TSAsExpression", node, opts);
***REMOVED***
function assertTSSatisfiesExpression(node, opts) {
  assert("TSSatisfiesExpression", node, opts);
***REMOVED***
function assertTSTypeAssertion(node, opts) {
  assert("TSTypeAssertion", node, opts);
***REMOVED***
function assertTSEnumBody(node, opts) {
  assert("TSEnumBody", node, opts);
***REMOVED***
function assertTSEnumDeclaration(node, opts) {
  assert("TSEnumDeclaration", node, opts);
***REMOVED***
function assertTSEnumMember(node, opts) {
  assert("TSEnumMember", node, opts);
***REMOVED***
function assertTSModuleDeclaration(node, opts) {
  assert("TSModuleDeclaration", node, opts);
***REMOVED***
function assertTSModuleBlock(node, opts) {
  assert("TSModuleBlock", node, opts);
***REMOVED***
function assertTSImportType(node, opts) {
  assert("TSImportType", node, opts);
***REMOVED***
function assertTSImportEqualsDeclaration(node, opts) {
  assert("TSImportEqualsDeclaration", node, opts);
***REMOVED***
function assertTSExternalModuleReference(node, opts) {
  assert("TSExternalModuleReference", node, opts);
***REMOVED***
function assertTSNonNullExpression(node, opts) {
  assert("TSNonNullExpression", node, opts);
***REMOVED***
function assertTSExportAssignment(node, opts) {
  assert("TSExportAssignment", node, opts);
***REMOVED***
function assertTSNamespaceExportDeclaration(node, opts) {
  assert("TSNamespaceExportDeclaration", node, opts);
***REMOVED***
function assertTSTypeAnnotation(node, opts) {
  assert("TSTypeAnnotation", node, opts);
***REMOVED***
function assertTSTypeParameterInstantiation(node, opts) {
  assert("TSTypeParameterInstantiation", node, opts);
***REMOVED***
function assertTSTypeParameterDeclaration(node, opts) {
  assert("TSTypeParameterDeclaration", node, opts);
***REMOVED***
function assertTSTypeParameter(node, opts) {
  assert("TSTypeParameter", node, opts);
***REMOVED***
function assertStandardized(node, opts) {
  assert("Standardized", node, opts);
***REMOVED***
function assertExpression(node, opts) {
  assert("Expression", node, opts);
***REMOVED***
function assertBinary(node, opts) {
  assert("Binary", node, opts);
***REMOVED***
function assertScopable(node, opts) {
  assert("Scopable", node, opts);
***REMOVED***
function assertBlockParent(node, opts) {
  assert("BlockParent", node, opts);
***REMOVED***
function assertBlock(node, opts) {
  assert("Block", node, opts);
***REMOVED***
function assertStatement(node, opts) {
  assert("Statement", node, opts);
***REMOVED***
function assertTerminatorless(node, opts) {
  assert("Terminatorless", node, opts);
***REMOVED***
function assertCompletionStatement(node, opts) {
  assert("CompletionStatement", node, opts);
***REMOVED***
function assertConditional(node, opts) {
  assert("Conditional", node, opts);
***REMOVED***
function assertLoop(node, opts) {
  assert("Loop", node, opts);
***REMOVED***
function assertWhile(node, opts) {
  assert("While", node, opts);
***REMOVED***
function assertExpressionWrapper(node, opts) {
  assert("ExpressionWrapper", node, opts);
***REMOVED***
function assertFor(node, opts) {
  assert("For", node, opts);
***REMOVED***
function assertForXStatement(node, opts) {
  assert("ForXStatement", node, opts);
***REMOVED***
function assertFunction(node, opts) {
  assert("Function", node, opts);
***REMOVED***
function assertFunctionParent(node, opts) {
  assert("FunctionParent", node, opts);
***REMOVED***
function assertPureish(node, opts) {
  assert("Pureish", node, opts);
***REMOVED***
function assertDeclaration(node, opts) {
  assert("Declaration", node, opts);
***REMOVED***
function assertPatternLike(node, opts) {
  assert("PatternLike", node, opts);
***REMOVED***
function assertLVal(node, opts) {
  assert("LVal", node, opts);
***REMOVED***
function assertTSEntityName(node, opts) {
  assert("TSEntityName", node, opts);
***REMOVED***
function assertLiteral(node, opts) {
  assert("Literal", node, opts);
***REMOVED***
function assertImmutable(node, opts) {
  assert("Immutable", node, opts);
***REMOVED***
function assertUserWhitespacable(node, opts) {
  assert("UserWhitespacable", node, opts);
***REMOVED***
function assertMethod(node, opts) {
  assert("Method", node, opts);
***REMOVED***
function assertObjectMember(node, opts) {
  assert("ObjectMember", node, opts);
***REMOVED***
function assertProperty(node, opts) {
  assert("Property", node, opts);
***REMOVED***
function assertUnaryLike(node, opts) {
  assert("UnaryLike", node, opts);
***REMOVED***
function assertPattern(node, opts) {
  assert("Pattern", node, opts);
***REMOVED***
function assertClass(node, opts) {
  assert("Class", node, opts);
***REMOVED***
function assertImportOrExportDeclaration(node, opts) {
  assert("ImportOrExportDeclaration", node, opts);
***REMOVED***
function assertExportDeclaration(node, opts) {
  assert("ExportDeclaration", node, opts);
***REMOVED***
function assertModuleSpecifier(node, opts) {
  assert("ModuleSpecifier", node, opts);
***REMOVED***
function assertAccessor(node, opts) {
  assert("Accessor", node, opts);
***REMOVED***
function assertPrivate(node, opts) {
  assert("Private", node, opts);
***REMOVED***
function assertFlow(node, opts) {
  assert("Flow", node, opts);
***REMOVED***
function assertFlowType(node, opts) {
  assert("FlowType", node, opts);
***REMOVED***
function assertFlowBaseAnnotation(node, opts) {
  assert("FlowBaseAnnotation", node, opts);
***REMOVED***
function assertFlowDeclaration(node, opts) {
  assert("FlowDeclaration", node, opts);
***REMOVED***
function assertFlowPredicate(node, opts) {
  assert("FlowPredicate", node, opts);
***REMOVED***
function assertEnumBody(node, opts) {
  assert("EnumBody", node, opts);
***REMOVED***
function assertEnumMember(node, opts) {
  assert("EnumMember", node, opts);
***REMOVED***
function assertJSX(node, opts) {
  assert("JSX", node, opts);
***REMOVED***
function assertMiscellaneous(node, opts) {
  assert("Miscellaneous", node, opts);
***REMOVED***
function assertTypeScript(node, opts) {
  assert("TypeScript", node, opts);
***REMOVED***
function assertTSTypeElement(node, opts) {
  assert("TSTypeElement", node, opts);
***REMOVED***
function assertTSType(node, opts) {
  assert("TSType", node, opts);
***REMOVED***
function assertTSBaseType(node, opts) {
  assert("TSBaseType", node, opts);
***REMOVED***
function assertNumberLiteral(node, opts) {
  (0, _deprecationWarning.default)("assertNumberLiteral", "assertNumericLiteral");
  assert("NumberLiteral", node, opts);
***REMOVED***
function assertRegexLiteral(node, opts) {
  (0, _deprecationWarning.default)("assertRegexLiteral", "assertRegExpLiteral");
  assert("RegexLiteral", node, opts);
***REMOVED***
function assertRestProperty(node, opts) {
  (0, _deprecationWarning.default)("assertRestProperty", "assertRestElement");
  assert("RestProperty", node, opts);
***REMOVED***
function assertSpreadProperty(node, opts) {
  (0, _deprecationWarning.default)("assertSpreadProperty", "assertSpreadElement");
  assert("SpreadProperty", node, opts);
***REMOVED***
function assertModuleDeclaration(node, opts) {
  (0, _deprecationWarning.default)("assertModuleDeclaration", "assertImportOrExportDeclaration");
  assert("ModuleDeclaration", node, opts);
***REMOVED***
