// jsx-ast-utils/src/getProp.js
import propName from './propName';

const DEFAULT_OPTIONS = {
  ignoreCase: true,
***REMOVED***;

/**
 * Returns the JSXAttribute itself or undefined, indicating the prop
 * is not present on the JSXOpeningElement.
 *
 */
export default function getProp(props = [], prop = '', options = DEFAULT_OPTIONS) {
  function getName(name) { return options.ignoreCase ? name.toUpperCase() : name; ***REMOVED***
  const propToFind = getName(prop);
  function isPropToFind(property) {
    return property.type === 'Property'
      && property.key.type === 'Identifier'
      && propToFind === getName(property.key.name);
  ***REMOVED***

  const foundAttribute = props.find((attribute) => {
    // If the props contain a spread prop, try to find the property in the object expression.
    if (attribute.type === 'JSXSpreadAttribute') {
      return attribute.argument.type === 'ObjectExpression'
        && propToFind !== getName('key') // https://github.com/reactjs/rfcs/pull/107
        && attribute.argument.properties.some(isPropToFind);
    ***REMOVED***

    return propToFind === getName(propName(attribute));
  ***REMOVED***);

  if (foundAttribute && foundAttribute.type === 'JSXSpreadAttribute') {
    return propertyToJSXAttribute(foundAttribute.argument.properties.find(isPropToFind));
  ***REMOVED***

  return foundAttribute;
***REMOVED***

function propertyToJSXAttribute(node) {
  const { key, value ***REMOVED*** = node;
  return {
    type: 'JSXAttribute',
    name: { type: 'JSXIdentifier', name: key.name, ...getBaseProps(key) ***REMOVED***,
    value: value.type === 'Literal'
      ? adjustRangeOfNode(value)
      : { type: 'JSXExpressionContainer', expression: adjustExpressionRange(value), ...getBaseProps(value) ***REMOVED***,
    ...getBaseProps(node),
  ***REMOVED***;
***REMOVED***

function adjustRangeOfNode(node) {
  const [start, end] = node.range || [node.start, node.end];

  return {
    ...node,
    end: undefined,
    range: [start, end],
    start: undefined,
  ***REMOVED***;
***REMOVED***

function adjustExpressionRange({ expressions, quasis, ...expression ***REMOVED***) {
  return {
    ...adjustRangeOfNode(expression),
    ...(expressions ? { expressions: expressions.map(adjustRangeOfNode) ***REMOVED*** : {***REMOVED***),
    ...(quasis ? { quasis: quasis.map(adjustRangeOfNode) ***REMOVED*** : {***REMOVED***),
  ***REMOVED***;
***REMOVED***

function getBaseProps({ loc, ...node ***REMOVED***) {
  const { range ***REMOVED*** = adjustRangeOfNode(node);

  return {
    loc: getBaseLocation(loc),
    range,
  ***REMOVED***;
***REMOVED***

function getBaseLocation({
  start,
  end,
  source,
  filename,
***REMOVED***) {
  return {
    start,
    end,
    ...(source !== undefined ? { source ***REMOVED*** : {***REMOVED***),
    ...(filename !== undefined ? { filename ***REMOVED*** : {***REMOVED***),
  ***REMOVED***;
***REMOVED***
