// lines-and-columns/build/index.js
"use strict";
exports.__esModule = true;
exports.LinesAndColumns = void 0;
var LF = '\n';
var CR = '\r';
var LinesAndColumns = /** @class */ (function () {
    function LinesAndColumns(string) {
        this.string = string;
        var offsets = [0];
        for (var offset = 0; offset < string.length;) {
            switch (string[offset]) {
                case LF:
                    offset += LF.length;
                    offsets.push(offset);
                    break;
                case CR:
                    offset += CR.length;
                    if (string[offset] === LF) {
                        offset += LF.length;
                    ***REMOVED***
                    offsets.push(offset);
                    break;
                default:
                    offset++;
                    break;
            ***REMOVED***
        ***REMOVED***
        this.offsets = offsets;
    ***REMOVED***
    LinesAndColumns.prototype.locationForIndex = function (index) {
        if (index < 0 || index > this.string.length) {
            return null;
        ***REMOVED***
        var line = 0;
        var offsets = this.offsets;
        while (offsets[line + 1] <= index) {
            line++;
        ***REMOVED***
        var column = index - offsets[line];
        return { line: line, column: column ***REMOVED***;
    ***REMOVED***;
    LinesAndColumns.prototype.indexForLocation = function (location) {
        var line = location.line, column = location.column;
        if (line < 0 || line >= this.offsets.length) {
            return null;
        ***REMOVED***
        if (column < 0 || column > this.lengthOfLine(line)) {
            return null;
        ***REMOVED***
        return this.offsets[line] + column;
    ***REMOVED***;
    LinesAndColumns.prototype.lengthOfLine = function (line) {
        var offset = this.offsets[line];
        var nextOffset = line === this.offsets.length - 1
            ? this.string.length
            : this.offsets[line + 1];
        return nextOffset - offset;
    ***REMOVED***;
    return LinesAndColumns;
***REMOVED***());
exports.LinesAndColumns = LinesAndColumns;
exports["default"] = LinesAndColumns;
