// @webassemblyjs/wasm-edit/esm/index.js
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {***REMOVED***; return { s: F, n: function n() { if (i >= o.length) return { done: true ***REMOVED***; return { done: false, value: o[i++] ***REMOVED***; ***REMOVED***, e: function e(_e) { throw _e; ***REMOVED***, f: F ***REMOVED***; ***REMOVED*** throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); ***REMOVED*** var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); ***REMOVED***, n: function n() { var step = it.next(); normalCompletion = step.done; return step; ***REMOVED***, e: function e(_e2) { didErr = true; err = _e2; ***REMOVED***, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); ***REMOVED*** finally { if (didErr) throw err; ***REMOVED*** ***REMOVED*** ***REMOVED***; ***REMOVED***

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); ***REMOVED***

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; ***REMOVED*** return arr2; ***REMOVED***

import { decode ***REMOVED*** from "@webassemblyjs/wasm-parser";
import { traverse ***REMOVED*** from "@webassemblyjs/ast";
import { cloneNode ***REMOVED*** from "@webassemblyjs/ast/lib/clone";
import { shrinkPaddedLEB128 ***REMOVED*** from "@webassemblyjs/wasm-opt";
import { getSectionForNode ***REMOVED*** from "@webassemblyjs/helper-wasm-bytecode";
import constants from "@webassemblyjs/helper-wasm-bytecode";
import { applyOperations ***REMOVED*** from "./apply";

function hashNode(node) {
  return JSON.stringify(node);
***REMOVED***

function preprocess(ab) {
  var optBin = shrinkPaddedLEB128(new Uint8Array(ab));
  return optBin.buffer;
***REMOVED***

function sortBySectionOrder(nodes) {
  var originalOrder = new Map();

  var _iterator = _createForOfIteratorHelper(nodes),
      _step;

  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var node = _step.value;
      originalOrder.set(node, originalOrder.size);
    ***REMOVED***
  ***REMOVED*** catch (err) {
    _iterator.e(err);
  ***REMOVED*** finally {
    _iterator.f();
  ***REMOVED***

  nodes.sort(function (a, b) {
    var sectionA = getSectionForNode(a);
    var sectionB = getSectionForNode(b);
    var aId = constants.sections[sectionA];
    var bId = constants.sections[sectionB];

    if (typeof aId !== "number" || typeof bId !== "number") {
      throw new Error("Section id not found");
    ***REMOVED***

    if (aId === bId) {
      // $FlowIgnore originalOrder is filled for all nodes
      return originalOrder.get(a) - originalOrder.get(b);
    ***REMOVED***

    return aId - bId;
  ***REMOVED***);
***REMOVED***

export function edit(ab, visitors) {
  ab = preprocess(ab);
  var ast = decode(ab);
  return editWithAST(ast, ab, visitors);
***REMOVED***
export function editWithAST(ast, ab, visitors) {
  var operations = [];
  var uint8Buffer = new Uint8Array(ab);
  var nodeBefore;

  function before(type, path) {
    nodeBefore = cloneNode(path.node);
  ***REMOVED***

  function after(type, path) {
    if (path.node._deleted === true) {
      operations.push({
        kind: "delete",
        node: path.node
      ***REMOVED***); // $FlowIgnore
    ***REMOVED*** else if (hashNode(nodeBefore) !== hashNode(path.node)) {
      operations.push({
        kind: "update",
        oldNode: nodeBefore,
        node: path.node
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***

  traverse(ast, visitors, before, after);
  uint8Buffer = applyOperations(ast, uint8Buffer, operations);
  return uint8Buffer.buffer;
***REMOVED***
export function add(ab, newNodes) {
  ab = preprocess(ab);
  var ast = decode(ab);
  return addWithAST(ast, ab, newNodes);
***REMOVED***
export function addWithAST(ast, ab, newNodes) {
  // Sort nodes by insertion order
  sortBySectionOrder(newNodes);
  var uint8Buffer = new Uint8Array(ab); // Map node into operations

  var operations = newNodes.map(function (n) {
    return {
      kind: "add",
      node: n
    ***REMOVED***;
  ***REMOVED***);
  uint8Buffer = applyOperations(ast, uint8Buffer, operations);
  return uint8Buffer.buffer;
***REMOVED***
