// minimist/index.js
'use strict';

function hasKey(obj, keys) {
	var o = obj;
	keys.slice(0, -1).forEach(function (key) {
		o = o[key] || {***REMOVED***;
	***REMOVED***);

	var key = keys[keys.length - 1];
	return key in o;
***REMOVED***

function isNumber(x) {
	if (typeof x === 'number') { return true; ***REMOVED***
	if ((/^0x[0-9a-f]+$/i).test(x)) { return true; ***REMOVED***
	return (/^[-+]?(?:\d+(?:\.\d*)?|\.\d+)(e[-+]?\d+)?$/).test(x);
***REMOVED***

function isConstructorOrProto(obj, key) {
	return (key === 'constructor' && typeof obj[key] === 'function') || key === '__proto__';
***REMOVED***

module.exports = function (args, opts) {
	if (!opts) { opts = {***REMOVED***; ***REMOVED***

	var flags = {
		bools: {***REMOVED***,
		strings: {***REMOVED***,
		unknownFn: null,
	***REMOVED***;

	if (typeof opts.unknown === 'function') {
		flags.unknownFn = opts.unknown;
	***REMOVED***

	if (typeof opts.boolean === 'boolean' && opts.boolean) {
		flags.allBools = true;
	***REMOVED*** else {
		[].concat(opts.boolean).filter(Boolean).forEach(function (key) {
			flags.bools[key] = true;
		***REMOVED***);
	***REMOVED***

	var aliases = {***REMOVED***;

	function aliasIsBoolean(key) {
		return aliases[key].some(function (x) {
			return flags.bools[x];
		***REMOVED***);
	***REMOVED***

	Object.keys(opts.alias || {***REMOVED***).forEach(function (key) {
		aliases[key] = [].concat(opts.alias[key]);
		aliases[key].forEach(function (x) {
			aliases[x] = [key].concat(aliases[key].filter(function (y) {
				return x !== y;
			***REMOVED***));
		***REMOVED***);
	***REMOVED***);

	[].concat(opts.string).filter(Boolean).forEach(function (key) {
		flags.strings[key] = true;
		if (aliases[key]) {
			[].concat(aliases[key]).forEach(function (k) {
				flags.strings[k] = true;
			***REMOVED***);
		***REMOVED***
	***REMOVED***);

	var defaults = opts.default || {***REMOVED***;

	var argv = { _: [] ***REMOVED***;

	function argDefined(key, arg) {
		return (flags.allBools && (/^--[^=]+$/).test(arg))
			|| flags.strings[key]
			|| flags.bools[key]
			|| aliases[key];
	***REMOVED***

	function setKey(obj, keys, value) {
		var o = obj;
		for (var i = 0; i < keys.length - 1; i++) {
			var key = keys[i];
			if (isConstructorOrProto(o, key)) { return; ***REMOVED***
			if (o[key] === undefined) { o[key] = {***REMOVED***; ***REMOVED***
			if (
				o[key] === Object.prototype
				|| o[key] === Number.prototype
				|| o[key] === String.prototype
			) {
				o[key] = {***REMOVED***;
			***REMOVED***
			if (o[key] === Array.prototype) { o[key] = []; ***REMOVED***
			o = o[key];
		***REMOVED***

		var lastKey = keys[keys.length - 1];
		if (isConstructorOrProto(o, lastKey)) { return; ***REMOVED***
		if (
			o === Object.prototype
			|| o === Number.prototype
			|| o === String.prototype
		) {
			o = {***REMOVED***;
		***REMOVED***
		if (o === Array.prototype) { o = []; ***REMOVED***
		if (o[lastKey] === undefined || flags.bools[lastKey] || typeof o[lastKey] === 'boolean') {
			o[lastKey] = value;
		***REMOVED*** else if (Array.isArray(o[lastKey])) {
			o[lastKey].push(value);
		***REMOVED*** else {
			o[lastKey] = [o[lastKey], value];
		***REMOVED***
	***REMOVED***

	function setArg(key, val, arg) {
		if (arg && flags.unknownFn && !argDefined(key, arg)) {
			if (flags.unknownFn(arg) === false) { return; ***REMOVED***
		***REMOVED***

		var value = !flags.strings[key] && isNumber(val)
			? Number(val)
			: val;
		setKey(argv, key.split('.'), value);

		(aliases[key] || []).forEach(function (x) {
			setKey(argv, x.split('.'), value);
		***REMOVED***);
	***REMOVED***

	Object.keys(flags.bools).forEach(function (key) {
		setArg(key, defaults[key] === undefined ? false : defaults[key]);
	***REMOVED***);

	var notFlags = [];

	if (args.indexOf('--') !== -1) {
		notFlags = args.slice(args.indexOf('--') + 1);
		args = args.slice(0, args.indexOf('--'));
	***REMOVED***

	for (var i = 0; i < args.length; i++) {
		var arg = args[i];
		var key;
		var next;

		if ((/^--.+=/).test(arg)) {
			// Using [\s\S] instead of . because js doesn't support the
			// 'dotall' regex modifier. See:
			// http://stackoverflow.com/a/1068308/13216
			var m = arg.match(/^--([^=]+)=([\s\S]*)$/);
			key = m[1];
			var value = m[2];
			if (flags.bools[key]) {
				value = value !== 'false';
			***REMOVED***
			setArg(key, value, arg);
		***REMOVED*** else if ((/^--no-.+/).test(arg)) {
			key = arg.match(/^--no-(.+)/)[1];
			setArg(key, false, arg);
		***REMOVED*** else if ((/^--.+/).test(arg)) {
			key = arg.match(/^--(.+)/)[1];
			next = args[i + 1];
			if (
				next !== undefined
				&& !(/^(-|--)[^-]/).test(next)
				&& !flags.bools[key]
				&& !flags.allBools
				&& (aliases[key] ? !aliasIsBoolean(key) : true)
			) {
				setArg(key, next, arg);
				i += 1;
			***REMOVED*** else if ((/^(true|false)$/).test(next)) {
				setArg(key, next === 'true', arg);
				i += 1;
			***REMOVED*** else {
				setArg(key, flags.strings[key] ? '' : true, arg);
			***REMOVED***
		***REMOVED*** else if ((/^-[^-]+/).test(arg)) {
			var letters = arg.slice(1, -1).split('');

			var broken = false;
			for (var j = 0; j < letters.length; j++) {
				next = arg.slice(j + 2);

				if (next === '-') {
					setArg(letters[j], next, arg);
					continue;
				***REMOVED***

				if ((/[A-Za-z]/).test(letters[j]) && next[0] === '=') {
					setArg(letters[j], next.slice(1), arg);
					broken = true;
					break;
				***REMOVED***

				if (
					(/[A-Za-z]/).test(letters[j])
					&& (/-?\d+(\.\d*)?(e-?\d+)?$/).test(next)
				) {
					setArg(letters[j], next, arg);
					broken = true;
					break;
				***REMOVED***

				if (letters[j + 1] && letters[j + 1].match(/\W/)) {
					setArg(letters[j], arg.slice(j + 2), arg);
					broken = true;
					break;
				***REMOVED*** else {
					setArg(letters[j], flags.strings[letters[j]] ? '' : true, arg);
				***REMOVED***
			***REMOVED***

			key = arg.slice(-1)[0];
			if (!broken && key !== '-') {
				if (
					args[i + 1]
					&& !(/^(-|--)[^-]/).test(args[i + 1])
					&& !flags.bools[key]
					&& (aliases[key] ? !aliasIsBoolean(key) : true)
				) {
					setArg(key, args[i + 1], arg);
					i += 1;
				***REMOVED*** else if (args[i + 1] && (/^(true|false)$/).test(args[i + 1])) {
					setArg(key, args[i + 1] === 'true', arg);
					i += 1;
				***REMOVED*** else {
					setArg(key, flags.strings[key] ? '' : true, arg);
				***REMOVED***
			***REMOVED***
		***REMOVED*** else {
			if (!flags.unknownFn || flags.unknownFn(arg) !== false) {
				argv._.push(flags.strings._ || !isNumber(arg) ? arg : Number(arg));
			***REMOVED***
			if (opts.stopEarly) {
				argv._.push.apply(argv._, args.slice(i + 1));
				break;
			***REMOVED***
		***REMOVED***
	***REMOVED***

	Object.keys(defaults).forEach(function (k) {
		if (!hasKey(argv, k.split('.'))) {
			setKey(argv, k.split('.'), defaults[k]);

			(aliases[k] || []).forEach(function (x) {
				setKey(argv, x.split('.'), defaults[k]);
			***REMOVED***);
		***REMOVED***
	***REMOVED***);

	if (opts['--']) {
		argv['--'] = notFlags.slice();
	***REMOVED*** else {
		notFlags.forEach(function (k) {
			argv._.push(k);
		***REMOVED***);
	***REMOVED***

	return argv;
***REMOVED***;
