// tailwindcss/src/featureFlags.js
import colors from 'picocolors'
import log from './util/log'

let defaults = {
  optimizeUniversalDefaults: false,
  generalizedModifiers: true,
  disableColorOpacityUtilitiesByDefault: false,
  relativeContentPathsByDefault: false,
***REMOVED***

let featureFlags = {
  future: [
    'hoverOnlyWhenSupported',
    'respectDefaultRingColorOpacity',
    'disableColorOpacityUtilitiesByDefault',
    'relativeContentPathsByDefault',
  ],
  experimental: ['optimizeUniversalDefaults', 'generalizedModifiers'],
***REMOVED***

export function flagEnabled(config, flag) {
  if (featureFlags.future.includes(flag)) {
    return config.future === 'all' || (config?.future?.[flag] ?? defaults[flag] ?? false)
  ***REMOVED***

  if (featureFlags.experimental.includes(flag)) {
    return (
      config.experimental === 'all' || (config?.experimental?.[flag] ?? defaults[flag] ?? false)
    )
  ***REMOVED***

  return false
***REMOVED***

function experimentalFlagsEnabled(config) {
  if (config.experimental === 'all') {
    return featureFlags.experimental
  ***REMOVED***

  return Object.keys(config?.experimental ?? {***REMOVED***).filter(
    (flag) => featureFlags.experimental.includes(flag) && config.experimental[flag]
  )
***REMOVED***

export function issueFlagNotices(config) {
  if (process.env.JEST_WORKER_ID !== undefined) {
    return
  ***REMOVED***

  if (experimentalFlagsEnabled(config).length > 0) {
    let changes = experimentalFlagsEnabled(config)
      .map((s) => colors.yellow(s))
      .join(', ')

    log.warn('experimental-flags-enabled', [
      `You have enabled experimental features: ${changes***REMOVED***`,
      'Experimental features in Tailwind CSS are not covered by semver, may introduce breaking changes, and can change at any time.',
    ])
  ***REMOVED***
***REMOVED***

export default featureFlags
