// webpack/lib/library/EnableLibraryPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

/** @typedef {import("../../declarations/WebpackOptions").LibraryOptions***REMOVED*** LibraryOptions */
/** @typedef {import("../../declarations/WebpackOptions").LibraryType***REMOVED*** LibraryType */
/** @typedef {import("../Compiler")***REMOVED*** Compiler */

/** @type {WeakMap<Compiler, Set<LibraryType>>***REMOVED*** */
const enabledTypes = new WeakMap();

/**
 * @typedef {object***REMOVED*** EnableLibraryPluginOptions
 * @property {() => void=***REMOVED*** additionalApply function that runs when applying the current plugin.
 */

/**
 * @param {Compiler***REMOVED*** compiler the compiler instance
 * @returns {Set<LibraryType>***REMOVED*** enabled types
 */
const getEnabledTypes = compiler => {
	let set = enabledTypes.get(compiler);
	if (set === undefined) {
		set = new Set();
		enabledTypes.set(compiler, set);
	***REMOVED***
	return set;
***REMOVED***;

class EnableLibraryPlugin {
	/**
	 * @param {LibraryType***REMOVED*** type library type that should be available
	 * @param {EnableLibraryPluginOptions***REMOVED*** options options of EnableLibraryPlugin
	 */
	constructor(type, options = {***REMOVED***) {
		/** @type {LibraryType***REMOVED*** */
		this.type = type;
		/** @type {EnableLibraryPluginOptions***REMOVED*** */
		this.options = options;
	***REMOVED***

	/**
	 * @param {Compiler***REMOVED*** compiler the compiler instance
	 * @param {LibraryType***REMOVED*** type type of library
	 * @returns {void***REMOVED***
	 */
	static setEnabled(compiler, type) {
		getEnabledTypes(compiler).add(type);
	***REMOVED***

	/**
	 * @param {Compiler***REMOVED*** compiler the compiler instance
	 * @param {LibraryType***REMOVED*** type type of library
	 * @returns {void***REMOVED***
	 */
	static checkEnabled(compiler, type) {
		if (!getEnabledTypes(compiler).has(type)) {
			throw new Error(
				`Library type "${type***REMOVED***" is not enabled. ` +
					"EnableLibraryPlugin need to be used to enable this type of library. " +
					'This usually happens through the "output.enabledLibraryTypes" option. ' +
					'If you are using a function as entry which sets "library", you need to add all potential library types to "output.enabledLibraryTypes". ' +
					`These types are enabled: ${Array.from(
						getEnabledTypes(compiler)
					).join(", ")***REMOVED***`
			);
		***REMOVED***
	***REMOVED***

	/**
	 * Apply the plugin
	 * @param {Compiler***REMOVED*** compiler the compiler instance
	 * @returns {void***REMOVED***
	 */
	apply(compiler) {
		const { type, options ***REMOVED*** = this;

		// Only enable once
		const enabled = getEnabledTypes(compiler);
		if (enabled.has(type)) return;
		enabled.add(type);

		if (typeof options.additionalApply === "function") {
			options.additionalApply();
		***REMOVED***

		if (typeof type === "string") {
			const enableExportProperty = () => {
				const ExportPropertyTemplatePlugin = require("./ExportPropertyLibraryPlugin");
				new ExportPropertyTemplatePlugin({
					type,
					nsObjectUsed: !["module", "modern-module"].includes(type),
					runtimeExportsUsed: !["module", "modern-module"].includes(type),
					renderStartupUsed: !["module", "modern-module"].includes(type)
				***REMOVED***).apply(compiler);
			***REMOVED***;
			switch (type) {
				case "var": {
					// @ts-expect-error https://github.com/microsoft/TypeScript/issues/41697
					const AssignLibraryPlugin = require("./AssignLibraryPlugin");
					new AssignLibraryPlugin({
						type,
						prefix: [],
						declare: "var",
						unnamed: "error"
					***REMOVED***).apply(compiler);
					break;
				***REMOVED***
				case "assign-properties": {
					// @ts-expect-error https://github.com/microsoft/TypeScript/issues/41697
					const AssignLibraryPlugin = require("./AssignLibraryPlugin");
					new AssignLibraryPlugin({
						type,
						prefix: [],
						declare: false,
						unnamed: "error",
						named: "copy"
					***REMOVED***).apply(compiler);
					break;
				***REMOVED***
				case "assign": {
					// @ts-expect-error https://github.com/microsoft/TypeScript/issues/41697
					const AssignLibraryPlugin = require("./AssignLibraryPlugin");
					new AssignLibraryPlugin({
						type,
						prefix: [],
						declare: false,
						unnamed: "error"
					***REMOVED***).apply(compiler);
					break;
				***REMOVED***
				case "this": {
					// @ts-expect-error https://github.com/microsoft/TypeScript/issues/41697
					const AssignLibraryPlugin = require("./AssignLibraryPlugin");
					new AssignLibraryPlugin({
						type,
						prefix: ["this"],
						declare: false,
						unnamed: "copy"
					***REMOVED***).apply(compiler);
					break;
				***REMOVED***
				case "window": {
					// @ts-expect-error https://github.com/microsoft/TypeScript/issues/41697
					const AssignLibraryPlugin = require("./AssignLibraryPlugin");
					new AssignLibraryPlugin({
						type,
						prefix: ["window"],
						declare: false,
						unnamed: "copy"
					***REMOVED***).apply(compiler);
					break;
				***REMOVED***
				case "self": {
					// @ts-expect-error https://github.com/microsoft/TypeScript/issues/41697
					const AssignLibraryPlugin = require("./AssignLibraryPlugin");
					new AssignLibraryPlugin({
						type,
						prefix: ["self"],
						declare: false,
						unnamed: "copy"
					***REMOVED***).apply(compiler);
					break;
				***REMOVED***
				case "global": {
					// @ts-expect-error https://github.com/microsoft/TypeScript/issues/41697
					const AssignLibraryPlugin = require("./AssignLibraryPlugin");
					new AssignLibraryPlugin({
						type,
						prefix: "global",
						declare: false,
						unnamed: "copy"
					***REMOVED***).apply(compiler);
					break;
				***REMOVED***
				case "commonjs": {
					// @ts-expect-error https://github.com/microsoft/TypeScript/issues/41697
					const AssignLibraryPlugin = require("./AssignLibraryPlugin");
					new AssignLibraryPlugin({
						type,
						prefix: ["exports"],
						declare: false,
						unnamed: "copy"
					***REMOVED***).apply(compiler);
					break;
				***REMOVED***
				case "commonjs-static": {
					// @ts-expect-error https://github.com/microsoft/TypeScript/issues/41697
					const AssignLibraryPlugin = require("./AssignLibraryPlugin");
					new AssignLibraryPlugin({
						type,
						prefix: ["exports"],
						declare: false,
						unnamed: "static"
					***REMOVED***).apply(compiler);
					break;
				***REMOVED***
				case "commonjs2":
				case "commonjs-module": {
					// @ts-expect-error https://github.com/microsoft/TypeScript/issues/41697
					const AssignLibraryPlugin = require("./AssignLibraryPlugin");
					new AssignLibraryPlugin({
						type,
						prefix: ["module", "exports"],
						declare: false,
						unnamed: "assign"
					***REMOVED***).apply(compiler);
					break;
				***REMOVED***
				case "amd":
				case "amd-require": {
					enableExportProperty();
					const AmdLibraryPlugin = require("./AmdLibraryPlugin");
					new AmdLibraryPlugin({
						type,
						requireAsWrapper: type === "amd-require"
					***REMOVED***).apply(compiler);
					break;
				***REMOVED***
				case "umd":
				case "umd2": {
					if (compiler.options.output.iife === false) {
						compiler.options.output.iife = true;

						class WarnFalseIifeUmdPlugin {
							/**
							 * @param {Compiler***REMOVED*** compiler the compiler instance
							 */
							apply(compiler) {
								compiler.hooks.thisCompilation.tap(
									"WarnFalseIifeUmdPlugin",
									compilation => {
										const FalseIIFEUmdWarning = require("../FalseIIFEUmdWarning");
										compilation.warnings.push(new FalseIIFEUmdWarning());
									***REMOVED***
								);
							***REMOVED***
						***REMOVED***

						new WarnFalseIifeUmdPlugin().apply(compiler);
					***REMOVED***
					enableExportProperty();
					const UmdLibraryPlugin = require("./UmdLibraryPlugin");
					new UmdLibraryPlugin({
						type,
						optionalAmdExternalAsGlobal: type === "umd2"
					***REMOVED***).apply(compiler);
					break;
				***REMOVED***
				case "system": {
					enableExportProperty();
					const SystemLibraryPlugin = require("./SystemLibraryPlugin");
					new SystemLibraryPlugin({
						type
					***REMOVED***).apply(compiler);
					break;
				***REMOVED***
				case "jsonp": {
					enableExportProperty();
					const JsonpLibraryPlugin = require("./JsonpLibraryPlugin");
					new JsonpLibraryPlugin({
						type
					***REMOVED***).apply(compiler);
					break;
				***REMOVED***
				case "module":
				case "modern-module": {
					enableExportProperty();
					const ModuleLibraryPlugin = require("./ModuleLibraryPlugin");
					new ModuleLibraryPlugin({
						type
					***REMOVED***).apply(compiler);
					break;
				***REMOVED***
				default:
					throw new Error(`Unsupported library type ${type***REMOVED***.
Plugins which provide custom library types must call EnableLibraryPlugin.setEnabled(compiler, type) to disable this error.`);
			***REMOVED***
		***REMOVED*** else {
			// TODO support plugin instances here
			// apply them to the compiler
		***REMOVED***
	***REMOVED***
***REMOVED***

module.exports = EnableLibraryPlugin;
