// jest-cli/build/init/questions.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
***REMOVED***);
exports.testScriptQuestion = exports.default = void 0;

/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const defaultQuestions = [
  {
    initial: false,
    message: 'Would you like to use Typescript for the configuration file?',
    name: 'useTypescript',
    type: 'confirm'
  ***REMOVED***,
  {
    choices: [
      {
        title: 'node',
        value: 'node'
      ***REMOVED***,
      {
        title: 'jsdom (browser-like)',
        value: 'jsdom'
      ***REMOVED***
    ],
    initial: 0,
    message: 'Choose the test environment that will be used for testing',
    name: 'environment',
    type: 'select'
  ***REMOVED***,
  {
    initial: false,
    message: 'Do you want Jest to add coverage reports?',
    name: 'coverage',
    type: 'confirm'
  ***REMOVED***,
  {
    choices: [
      {
        title: 'v8',
        value: 'v8'
      ***REMOVED***,
      {
        title: 'babel',
        value: 'babel'
      ***REMOVED***
    ],
    initial: 0,
    message: 'Which provider should be used to instrument code for coverage?',
    name: 'coverageProvider',
    type: 'select'
  ***REMOVED***,
  {
    initial: false,
    message:
      'Automatically clear mock calls, instances and results before every test?',
    name: 'clearMocks',
    type: 'confirm'
  ***REMOVED***
];
var _default = defaultQuestions;
exports.default = _default;
const testScriptQuestion = {
  initial: true,
  message:
    'Would you like to use Jest when running "test" script in "package.json"?',
  name: 'scripts',
  type: 'confirm'
***REMOVED***;
exports.testScriptQuestion = testScriptQuestion;
