// js-yaml/lib/js-yaml/schema.js
'use strict';

/*eslint-disable max-len*/

var common        = require('./common');
var YAMLException = require('./exception');
var Type          = require('./type');


function compileList(schema, name, result) {
  var exclude = [];

  schema.include.forEach(function (includedSchema) {
    result = compileList(includedSchema, name, result);
  ***REMOVED***);

  schema[name].forEach(function (currentType) {
    result.forEach(function (previousType, previousIndex) {
      if (previousType.tag === currentType.tag && previousType.kind === currentType.kind) {
        exclude.push(previousIndex);
      ***REMOVED***
    ***REMOVED***);

    result.push(currentType);
  ***REMOVED***);

  return result.filter(function (type, index) {
    return exclude.indexOf(index) === -1;
  ***REMOVED***);
***REMOVED***


function compileMap(/* lists... */) {
  var result = {
        scalar: {***REMOVED***,
        sequence: {***REMOVED***,
        mapping: {***REMOVED***,
        fallback: {***REMOVED***
      ***REMOVED***, index, length;

  function collectType(type) {
    result[type.kind][type.tag] = result['fallback'][type.tag] = type;
  ***REMOVED***

  for (index = 0, length = arguments.length; index < length; index += 1) {
    arguments[index].forEach(collectType);
  ***REMOVED***
  return result;
***REMOVED***


function Schema(definition) {
  this.include  = definition.include  || [];
  this.implicit = definition.implicit || [];
  this.explicit = definition.explicit || [];

  this.implicit.forEach(function (type) {
    if (type.loadKind && type.loadKind !== 'scalar') {
      throw new YAMLException('There is a non-scalar type in the implicit list of a schema. Implicit resolving of such types is not supported.');
    ***REMOVED***
  ***REMOVED***);

  this.compiledImplicit = compileList(this, 'implicit', []);
  this.compiledExplicit = compileList(this, 'explicit', []);
  this.compiledTypeMap  = compileMap(this.compiledImplicit, this.compiledExplicit);
***REMOVED***


Schema.DEFAULT = null;


Schema.create = function createSchema() {
  var schemas, types;

  switch (arguments.length) {
    case 1:
      schemas = Schema.DEFAULT;
      types = arguments[0];
      break;

    case 2:
      schemas = arguments[0];
      types = arguments[1];
      break;

    default:
      throw new YAMLException('Wrong number of arguments for Schema.create function');
  ***REMOVED***

  schemas = common.toArray(schemas);
  types = common.toArray(types);

  if (!schemas.every(function (schema) { return schema instanceof Schema; ***REMOVED***)) {
    throw new YAMLException('Specified list of super schemas (or a single Schema object) contains a non-Schema object.');
  ***REMOVED***

  if (!types.every(function (type) { return type instanceof Type; ***REMOVED***)) {
    throw new YAMLException('Specified list of YAML types (or a single Type object) contains a non-Type object.');
  ***REMOVED***

  return new Schema({
    include: schemas,
    explicit: types
  ***REMOVED***);
***REMOVED***;


module.exports = Schema;
