// @babel/helper-string-parser/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.readCodePoint = readCodePoint;
exports.readInt = readInt;
exports.readStringContents = readStringContents;
var _isDigit = function isDigit(code) {
  return code >= 48 && code <= 57;
***REMOVED***;
const forbiddenNumericSeparatorSiblings = {
  decBinOct: new Set([46, 66, 69, 79, 95, 98, 101, 111]),
  hex: new Set([46, 88, 95, 120])
***REMOVED***;
const isAllowedNumericSeparatorSibling = {
  bin: ch => ch === 48 || ch === 49,
  oct: ch => ch >= 48 && ch <= 55,
  dec: ch => ch >= 48 && ch <= 57,
  hex: ch => ch >= 48 && ch <= 57 || ch >= 65 && ch <= 70 || ch >= 97 && ch <= 102
***REMOVED***;
function readStringContents(type, input, pos, lineStart, curLine, errors) {
  const initialPos = pos;
  const initialLineStart = lineStart;
  const initialCurLine = curLine;
  let out = "";
  let firstInvalidLoc = null;
  let chunkStart = pos;
  const {
    length
  ***REMOVED*** = input;
  for (;;) {
    if (pos >= length) {
      errors.unterminated(initialPos, initialLineStart, initialCurLine);
      out += input.slice(chunkStart, pos);
      break;
    ***REMOVED***
    const ch = input.charCodeAt(pos);
    if (isStringEnd(type, ch, input, pos)) {
      out += input.slice(chunkStart, pos);
      break;
    ***REMOVED***
    if (ch === 92) {
      out += input.slice(chunkStart, pos);
      const res = readEscapedChar(input, pos, lineStart, curLine, type === "template", errors);
      if (res.ch === null && !firstInvalidLoc) {
        firstInvalidLoc = {
          pos,
          lineStart,
          curLine
        ***REMOVED***;
      ***REMOVED*** else {
        out += res.ch;
      ***REMOVED***
      ({
        pos,
        lineStart,
        curLine
      ***REMOVED*** = res);
      chunkStart = pos;
    ***REMOVED*** else if (ch === 8232 || ch === 8233) {
      ++pos;
      ++curLine;
      lineStart = pos;
    ***REMOVED*** else if (ch === 10 || ch === 13) {
      if (type === "template") {
        out += input.slice(chunkStart, pos) + "\n";
        ++pos;
        if (ch === 13 && input.charCodeAt(pos) === 10) {
          ++pos;
        ***REMOVED***
        ++curLine;
        chunkStart = lineStart = pos;
      ***REMOVED*** else {
        errors.unterminated(initialPos, initialLineStart, initialCurLine);
      ***REMOVED***
    ***REMOVED*** else {
      ++pos;
    ***REMOVED***
  ***REMOVED***
  return {
    pos,
    str: out,
    firstInvalidLoc,
    lineStart,
    curLine,
    containsInvalid: !!firstInvalidLoc
  ***REMOVED***;
***REMOVED***
function isStringEnd(type, ch, input, pos) {
  if (type === "template") {
    return ch === 96 || ch === 36 && input.charCodeAt(pos + 1) === 123;
  ***REMOVED***
  return ch === (type === "double" ? 34 : 39);
***REMOVED***
function readEscapedChar(input, pos, lineStart, curLine, inTemplate, errors) {
  const throwOnInvalid = !inTemplate;
  pos++;
  const res = ch => ({
    pos,
    ch,
    lineStart,
    curLine
  ***REMOVED***);
  const ch = input.charCodeAt(pos++);
  switch (ch) {
    case 110:
      return res("\n");
    case 114:
      return res("\r");
    case 120:
      {
        let code;
        ({
          code,
          pos
        ***REMOVED*** = readHexChar(input, pos, lineStart, curLine, 2, false, throwOnInvalid, errors));
        return res(code === null ? null : String.fromCharCode(code));
      ***REMOVED***
    case 117:
      {
        let code;
        ({
          code,
          pos
        ***REMOVED*** = readCodePoint(input, pos, lineStart, curLine, throwOnInvalid, errors));
        return res(code === null ? null : String.fromCodePoint(code));
      ***REMOVED***
    case 116:
      return res("\t");
    case 98:
      return res("\b");
    case 118:
      return res("\u000b");
    case 102:
      return res("\f");
    case 13:
      if (input.charCodeAt(pos) === 10) {
        ++pos;
      ***REMOVED***
    case 10:
      lineStart = pos;
      ++curLine;
    case 8232:
    case 8233:
      return res("");
    case 56:
    case 57:
      if (inTemplate) {
        return res(null);
      ***REMOVED*** else {
        errors.strictNumericEscape(pos - 1, lineStart, curLine);
      ***REMOVED***
    default:
      if (ch >= 48 && ch <= 55) {
        const startPos = pos - 1;
        const match = /^[0-7]+/.exec(input.slice(startPos, pos + 2));
        let octalStr = match[0];
        let octal = parseInt(octalStr, 8);
        if (octal > 255) {
          octalStr = octalStr.slice(0, -1);
          octal = parseInt(octalStr, 8);
        ***REMOVED***
        pos += octalStr.length - 1;
        const next = input.charCodeAt(pos);
        if (octalStr !== "0" || next === 56 || next === 57) {
          if (inTemplate) {
            return res(null);
          ***REMOVED*** else {
            errors.strictNumericEscape(startPos, lineStart, curLine);
          ***REMOVED***
        ***REMOVED***
        return res(String.fromCharCode(octal));
      ***REMOVED***
      return res(String.fromCharCode(ch));
  ***REMOVED***
***REMOVED***
function readHexChar(input, pos, lineStart, curLine, len, forceLen, throwOnInvalid, errors) {
  const initialPos = pos;
  let n;
  ({
    n,
    pos
  ***REMOVED*** = readInt(input, pos, lineStart, curLine, 16, len, forceLen, false, errors, !throwOnInvalid));
  if (n === null) {
    if (throwOnInvalid) {
      errors.invalidEscapeSequence(initialPos, lineStart, curLine);
    ***REMOVED*** else {
      pos = initialPos - 1;
    ***REMOVED***
  ***REMOVED***
  return {
    code: n,
    pos
  ***REMOVED***;
***REMOVED***
function readInt(input, pos, lineStart, curLine, radix, len, forceLen, allowNumSeparator, errors, bailOnError) {
  const start = pos;
  const forbiddenSiblings = radix === 16 ? forbiddenNumericSeparatorSiblings.hex : forbiddenNumericSeparatorSiblings.decBinOct;
  const isAllowedSibling = radix === 16 ? isAllowedNumericSeparatorSibling.hex : radix === 10 ? isAllowedNumericSeparatorSibling.dec : radix === 8 ? isAllowedNumericSeparatorSibling.oct : isAllowedNumericSeparatorSibling.bin;
  let invalid = false;
  let total = 0;
  for (let i = 0, e = len == null ? Infinity : len; i < e; ++i) {
    const code = input.charCodeAt(pos);
    let val;
    if (code === 95 && allowNumSeparator !== "bail") {
      const prev = input.charCodeAt(pos - 1);
      const next = input.charCodeAt(pos + 1);
      if (!allowNumSeparator) {
        if (bailOnError) return {
          n: null,
          pos
        ***REMOVED***;
        errors.numericSeparatorInEscapeSequence(pos, lineStart, curLine);
      ***REMOVED*** else if (Number.isNaN(next) || !isAllowedSibling(next) || forbiddenSiblings.has(prev) || forbiddenSiblings.has(next)) {
        if (bailOnError) return {
          n: null,
          pos
        ***REMOVED***;
        errors.unexpectedNumericSeparator(pos, lineStart, curLine);
      ***REMOVED***
      ++pos;
      continue;
    ***REMOVED***
    if (code >= 97) {
      val = code - 97 + 10;
    ***REMOVED*** else if (code >= 65) {
      val = code - 65 + 10;
    ***REMOVED*** else if (_isDigit(code)) {
      val = code - 48;
    ***REMOVED*** else {
      val = Infinity;
    ***REMOVED***
    if (val >= radix) {
      if (val <= 9 && bailOnError) {
        return {
          n: null,
          pos
        ***REMOVED***;
      ***REMOVED*** else if (val <= 9 && errors.invalidDigit(pos, lineStart, curLine, radix)) {
        val = 0;
      ***REMOVED*** else if (forceLen) {
        val = 0;
        invalid = true;
      ***REMOVED*** else {
        break;
      ***REMOVED***
    ***REMOVED***
    ++pos;
    total = total * radix + val;
  ***REMOVED***
  if (pos === start || len != null && pos - start !== len || invalid) {
    return {
      n: null,
      pos
    ***REMOVED***;
  ***REMOVED***
  return {
    n: total,
    pos
  ***REMOVED***;
***REMOVED***
function readCodePoint(input, pos, lineStart, curLine, throwOnInvalid, errors) {
  const ch = input.charCodeAt(pos);
  let code;
  if (ch === 123) {
    ++pos;
    ({
      code,
      pos
    ***REMOVED*** = readHexChar(input, pos, lineStart, curLine, input.indexOf("***REMOVED***", pos) - pos, true, throwOnInvalid, errors));
    ++pos;
    if (code !== null && code > 0x10ffff) {
      if (throwOnInvalid) {
        errors.invalidCodePoint(pos, lineStart, curLine);
      ***REMOVED*** else {
        return {
          code: null,
          pos
        ***REMOVED***;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED*** else {
    ({
      code,
      pos
    ***REMOVED*** = readHexChar(input, pos, lineStart, curLine, 4, false, throwOnInvalid, errors));
  ***REMOVED***
  return {
    code,
    pos
  ***REMOVED***;
***REMOVED***
