// check-types/src/check-types.js
/*globals define, module, Symbol, Map, Set */
/*jshint -W056 */

(function (globals) {
  'use strict';

  var messages, predicates, functions, assert, not, maybe, collections,
     hasOwnProperty, toString, keys, slice, isArray, neginf, posinf,
     haveSymbols, haveMaps, haveSets;

  messages = {***REMOVED***;
  predicates = {***REMOVED***;

  [
    { n: 'equal', f: equal, s: 'equal {e***REMOVED***' ***REMOVED***,
    { n: 'undefined', f: isUndefined, s: 'be undefined' ***REMOVED***,
    { n: 'null', f: isNull, s: 'be null' ***REMOVED***,
    { n: 'assigned', f: assigned, s: 'be assigned' ***REMOVED***,
    { n: 'primitive', f: primitive, s: 'be primitive type' ***REMOVED***,
    { n: 'contains', f: contains, s: 'contain {e***REMOVED***' ***REMOVED***,
    { n: 'in', f: isIn, s: 'be in {e***REMOVED***' ***REMOVED***,
    { n: 'containsKey', f: containsKey, s: 'contain key {e***REMOVED***' ***REMOVED***,
    { n: 'keyIn', f: keyIn, s: 'be key in {e***REMOVED***' ***REMOVED***,
    { n: 'zero', f: zero, s: 'be 0' ***REMOVED***,
    { n: 'one', f: one, s: 'be 1' ***REMOVED***,
    { n: 'infinity', f: infinity, s: 'be infinity' ***REMOVED***,
    { n: 'number', f: number, s: 'be Number' ***REMOVED***,
    { n: 'integer', f: integer, s: 'be integer' ***REMOVED***,
    { n: 'float', f: float, s: 'be non-integer number' ***REMOVED***,
    { n: 'even', f: even, s: 'be even number' ***REMOVED***,
    { n: 'odd', f: odd, s: 'be odd number' ***REMOVED***,
    { n: 'greater', f: greater, s: 'be greater than {e***REMOVED***' ***REMOVED***,
    { n: 'less', f: less, s: 'be less than {e***REMOVED***' ***REMOVED***,
    { n: 'between', f: between, s: 'be between {e***REMOVED*** and {e2***REMOVED***' ***REMOVED***,
    { n: 'greaterOrEqual', f: greaterOrEqual, s: 'be greater than or equal to {e***REMOVED***' ***REMOVED***,
    { n: 'lessOrEqual', f: lessOrEqual, s: 'be less than or equal to {e***REMOVED***' ***REMOVED***,
    { n: 'inRange', f: inRange, s: 'be in the range {e***REMOVED*** to {e2***REMOVED***' ***REMOVED***,
    { n: 'positive', f: positive, s: 'be positive number' ***REMOVED***,
    { n: 'negative', f: negative, s: 'be negative number' ***REMOVED***,
    { n: 'string', f: string, s: 'be String' ***REMOVED***,
    { n: 'emptyString', f: emptyString, s: 'be empty string' ***REMOVED***,
    { n: 'nonEmptyString', f: nonEmptyString, s: 'be non-empty string' ***REMOVED***,
    { n: 'match', f: match, s: 'match {e***REMOVED***' ***REMOVED***,
    { n: 'boolean', f: boolean, s: 'be Boolean' ***REMOVED***,
    { n: 'object', f: object, s: 'be Object' ***REMOVED***,
    { n: 'emptyObject', f: emptyObject, s: 'be empty object' ***REMOVED***,
    { n: 'nonEmptyObject', f: nonEmptyObject, s: 'be non-empty object' ***REMOVED***,
    { n: 'instanceStrict', f: instanceStrict, s: 'be instanceof {t***REMOVED***' ***REMOVED***,
    { n: 'thenable', f: thenable, s: 'be promise-like' ***REMOVED***,
    { n: 'instance', f: instance, s: 'be {t***REMOVED***' ***REMOVED***,
    { n: 'like', f: like, s: 'be like {e***REMOVED***' ***REMOVED***,
    { n: 'identical', f: identical, s: 'be identical to {e***REMOVED***' ***REMOVED***,
    { n: 'array', f: array, s: 'be Array' ***REMOVED***,
    { n: 'emptyArray', f: emptyArray, s: 'be empty array' ***REMOVED***,
    { n: 'nonEmptyArray', f: nonEmptyArray, s: 'be non-empty array' ***REMOVED***,
    { n: 'arrayLike', f: arrayLike, s: 'be array-like' ***REMOVED***,
    { n: 'iterable', f: iterable, s: 'be iterable' ***REMOVED***,
    { n: 'date', f: date, s: 'be valid Date' ***REMOVED***,
    { n: 'function', f: isFunction, s: 'be Function' ***REMOVED***,
    { n: 'hasLength', f: hasLength, s: 'have length {e***REMOVED***' ***REMOVED***,
    { n: 'throws', f: throws, s: 'throw' ***REMOVED***
  ].map(function (data) {
    var n = data.n;
    messages[n] = 'assert failed: expected {a***REMOVED*** to ' + data.s;
    predicates[n] = data.f;
  ***REMOVED***);

  functions = {
    map: map,
    all: all,
    any: any
  ***REMOVED***;

  collections = [ 'array', 'arrayLike', 'iterable', 'object' ];
  hasOwnProperty = Object.prototype.hasOwnProperty;
  toString = Object.prototype.toString;
  keys = Object.keys;
  slice = Array.prototype.slice;
  isArray = Array.isArray;
  neginf = Number.NEGATIVE_INFINITY;
  posinf = Number.POSITIVE_INFINITY;
  haveSymbols = typeof Symbol === 'function';
  haveMaps = typeof Map === 'function';
  haveSets = typeof Set === 'function';

  functions = mixin(functions, predicates);
  assert = createModifiedPredicates(assertModifier, assertImpl);
  not = createModifiedPredicates(notModifier, notImpl);
  maybe = createModifiedPredicates(maybeModifier, maybeImpl);
  assert.not = createModifiedModifier(assertModifier, not, 'not ');
  assert.maybe = createModifiedModifier(assertModifier, maybe, 'maybe ');

  collections.forEach(createOfPredicates);
  createOfModifiers(assert, assertModifier);
  createOfModifiers(not, notModifier);
  collections.forEach(createMaybeOfModifiers);

  exportFunctions(mixin(functions, {
    assert: assert,
    not: not,
    maybe: maybe
  ***REMOVED***));

  /**
   * Public function `equal`.
   *
   * Returns true if `lhs` and `rhs` are strictly equal, without coercion.
   * Returns false otherwise.
   */
  function equal (lhs, rhs) {
    return lhs === rhs;
  ***REMOVED***

  /**
   * Public function `undefined`.
   *
   * Returns true if `data` is undefined, false otherwise.
   */
  function isUndefined (data) {
    return data === undefined;
  ***REMOVED***

  /**
   * Public function `null`.
   *
   * Returns true if `data` is null, false otherwise.
   */
  function isNull (data) {
    return data === null;
  ***REMOVED***

  /**
   * Public function `assigned`.
   *
   * Returns true if `data` is not null or undefined, false otherwise.
   */
  function assigned (data) {
    return data !== undefined && data !== null;
  ***REMOVED***

  /**
   * Public function `primitive`.
   *
   * Returns true if `data` is a primitive type, false otherwise.
   */
  function primitive (data) {
    var type;

    switch (data) {
      case null:
      case undefined:
      case false:
      case true:
        return true;
    ***REMOVED***

    type = typeof data;
    return type === 'string' || type === 'number' || (haveSymbols && type === 'symbol');
  ***REMOVED***

  /**
   * Public function `zero`.
   *
   * Returns true if `data` is zero, false otherwise.
   */
  function zero (data) {
    return data === 0;
  ***REMOVED***

  /**
   * Public function `one`.
   *
   * Returns true if `data` is one, false otherwise.
   */
  function one (data) {
    return data === 1;
  ***REMOVED***

  /**
   * Public function `infinity`.
   *
   * Returns true if `data` is positive or negative infinity, false otherwise.
   */
  function infinity (data) {
    return data === neginf || data === posinf;
  ***REMOVED***

  /**
   * Public function `number`.
   *
   * Returns true if `data` is a number, false otherwise.
   */
  function number (data) {
    return typeof data === 'number' && data > neginf && data < posinf;
  ***REMOVED***

  /**
   * Public function `integer`.
   *
   * Returns true if `data` is an integer, false otherwise.
   */
  function integer (data) {
    return typeof data === 'number' && data % 1 === 0;
  ***REMOVED***

  /**
   * Public function `float`.
   *
   * Returns true if `data` is a non-integer number, false otherwise.
   */
  function float (data) {
    return number(data) && data % 1 !== 0;
  ***REMOVED***

  /**
   * Public function `even`.
   *
   * Returns true if `data` is an even number, false otherwise.
   */
  function even (data) {
    return typeof data === 'number' && data % 2 === 0;
  ***REMOVED***

  /**
   * Public function `odd`.
   *
   * Returns true if `data` is an odd number, false otherwise.
   */
  function odd (data) {
    return integer(data) && data % 2 !== 0;
  ***REMOVED***

  /**
   * Public function `greater`.
   *
   * Returns true if `lhs` is a number greater than `rhs`, false otherwise.
   */
  function greater (lhs, rhs) {
    return number(lhs) && lhs > rhs;
  ***REMOVED***

  /**
   * Public function `less`.
   *
   * Returns true if `lhs` is a number less than `rhs`, false otherwise.
   */
  function less (lhs, rhs) {
    return number(lhs) && lhs < rhs;
  ***REMOVED***

  /**
   * Public function `between`.
   *
   * Returns true if `data` is a number between `x` and `y`, false otherwise.
   */
  function between (data, x, y) {
    if (x < y) {
      return greater(data, x) && data < y;
    ***REMOVED***

    return less(data, x) && data > y;
  ***REMOVED***

  /**
   * Public function `greaterOrEqual`.
   *
   * Returns true if `lhs` is a number greater than or equal to `rhs`, false
   * otherwise.
   */
  function greaterOrEqual (lhs, rhs) {
    return number(lhs) && lhs >= rhs;
  ***REMOVED***

  /**
   * Public function `lessOrEqual`.
   *
   * Returns true if `lhs` is a number less than or equal to `rhs`, false
   * otherwise.
   */
  function lessOrEqual (lhs, rhs) {
    return number(lhs) && lhs <= rhs;
  ***REMOVED***

  /**
   * Public function `inRange`.
   *
   * Returns true if `data` is a number in the range `x..y`, false otherwise.
   */
  function inRange (data, x, y) {
    if (x < y) {
      return greaterOrEqual(data, x) && data <= y;
    ***REMOVED***

    return lessOrEqual(data, x) && data >= y;
  ***REMOVED***

  /**
   * Public function `positive`.
   *
   * Returns true if `data` is a positive number, false otherwise.
   */
  function positive (data) {
    return greater(data, 0);
  ***REMOVED***

  /**
   * Public function `negative`.
   *
   * Returns true if `data` is a negative number, false otherwise.
   */
  function negative (data) {
    return less(data, 0);
  ***REMOVED***

  /**
   * Public function `string`.
   *
   * Returns true if `data` is a string, false otherwise.
   */
  function string (data) {
    return typeof data === 'string';
  ***REMOVED***

  /**
   * Public function `emptyString`.
   *
   * Returns true if `data` is the empty string, false otherwise.
   */
  function emptyString (data) {
    return data === '';
  ***REMOVED***

  /**
   * Public function `nonEmptyString`.
   *
   * Returns true if `data` is a non-empty string, false otherwise.
   */
  function nonEmptyString (data) {
    return string(data) && data !== '';
  ***REMOVED***

  /**
   * Public function `match`.
   *
   * Returns true if `data` is a string that matches `regex`, false otherwise.
   */
  function match (data, regex) {
    return string(data) && !! data.match(regex);
  ***REMOVED***

  /**
   * Public function `boolean`.
   *
   * Returns true if `data` is a boolean value, false otherwise.
   */
  function boolean (data) {
    return data === false || data === true;
  ***REMOVED***

  /**
   * Public function `object`.
   *
   * Returns true if `data` is a plain-old JS object, false otherwise.
   */
  function object (data) {
    return toString.call(data) === '[object Object]';
  ***REMOVED***

  /**
   * Public function `emptyObject`.
   *
   * Returns true if `data` is an empty object, false otherwise.
   */
  function emptyObject (data) {
    return object(data) && ! some(data, function () {
      return true;
    ***REMOVED***);
  ***REMOVED***

  function some (data, predicate) {
    for (var key in data) {
      if (hasOwnProperty.call(data, key)) {
        if (predicate(key, data[key])) {
          return true;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    return false;
  ***REMOVED***

  /**
   * Public function `nonEmptyObject`.
   *
   * Returns true if `data` is a non-empty object, false otherwise.
   */
  function nonEmptyObject (data) {
    return object(data) && some(data, function () {
      return true;
    ***REMOVED***);
  ***REMOVED***

  /**
   * Public function `thenable`.
   *
   * Returns true if `data` has a `then` method.
   */
  function thenable (data) {
    return assigned(data) && isFunction(data.then);
  ***REMOVED***

  /**
   * Public function `instanceStrict`.
   *
   * Returns true if `data` is an instance of `prototype`, false otherwise.
   */
  function instanceStrict (data, prototype) {
    try {
      return data instanceof prototype;
    ***REMOVED*** catch (error) {
      return false;
    ***REMOVED***
  ***REMOVED***

  /**
   * Public function `instance`.
   *
   * Returns true if `data` is an instance of `prototype`, false otherwise.
   * Falls back to testing constructor.name and Object.prototype.toString
   * if the initial instanceof test fails.
   */
  function instance (data, prototype) {
    try {
      return instanceStrict(data, prototype) ||
        data.constructor.name === prototype.name ||
        toString.call(data) === '[object ' + prototype.name + ']';
    ***REMOVED*** catch (error) {
      return false;
    ***REMOVED***
  ***REMOVED***

  /**
   * Public function `like`.
   *
   * Tests whether `data` 'quacks like a duck'. Returns true if `data` has all
   * of the properties of `archetype` (the 'duck'), false otherwise. Interrogates
   * objects recursively, to arbitrary depth.
   */
  function like (data, archetype) {
    var name;

    if (! assigned(data) || ! assigned(archetype)) {
      return data === archetype;
    ***REMOVED***

    for (name in archetype) {
      if (hasOwnProperty.call(archetype, name)) {
        if (! hasOwnProperty.call(data, name) || typeof data[name] !== typeof archetype[name]) {
          return false;
        ***REMOVED***

        if (object(data[name]) && ! like(data[name], archetype[name])) {
          return false;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    return true;
  ***REMOVED***

  /**
   * Public function `identical`.
   *
   * Tests whether `data` has all of the same properties and values as `archetype`.
   * Interrogates objects recursively, to arbitrary depth.
   */
  function identical (data, archetype) {
    var name;

    if (! assigned(data) || ! assigned(archetype)) {
      return data === archetype;
    ***REMOVED***

    for (name in archetype) {
      if (hasOwnProperty.call(archetype, name)) {
        if (! hasOwnProperty.call(data, name)) {
          return false;
        ***REMOVED***

        if (object(data[name])) {
          if (! identical(data[name], archetype[name])) {
            return false;
          ***REMOVED***
        ***REMOVED*** else if (data[name] !== archetype[name]) {
          return false;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    for (name in data) {
      if (hasOwnProperty.call(data, name)) {
        if (! hasOwnProperty.call(archetype, name)) {
          return false;
        ***REMOVED***

        if (object(archetype[name])) {
          if (! identical(archetype[name], data[name])) {
            return false;
          ***REMOVED***
        ***REMOVED*** else if (archetype[name] !== data[name]) {
          return false;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    return true;
  ***REMOVED***

  /**
   * Public function `array`.
   *
   * Returns true if `data` is an array, false otherwise.
   */
  function array (data) {
    return isArray(data);
  ***REMOVED***

  /**
   * Public function `emptyArray`.
   *
   * Returns true if `data` is an empty array, false otherwise.
   */
  function emptyArray (data) {
    return isArray(data) && data.length === 0;
  ***REMOVED***

  /**
   * Public function `nonEmptyArray`.
   *
   * Returns true if `data` is a non-empty array, false otherwise.
   */
  function nonEmptyArray (data) {
    return isArray(data) && data.length > 0;
  ***REMOVED***

  /**
   * Public function `arrayLike`.
   *
   * Returns true if `data` is an array-like object, false otherwise.
   */
  function arrayLike (data) {
    return assigned(data) && data.length >= 0;
  ***REMOVED***

  /**
   * Public function `iterable`.
   *
   * Returns true if `data` is an iterable, false otherwise.
   */
  function iterable (data) {
    if (! haveSymbols) {
      // Fall back to `arrayLike` predicate in pre-ES6 environments.
      return arrayLike(data);
    ***REMOVED***

    return assigned(data) && isFunction(data[Symbol.iterator]);
  ***REMOVED***

  /**
   * Public function `contains`.
   *
   * Returns true if `data` contains `value`, false otherwise.
   * Works with objects, arrays and array-likes (including strings).
   */
  function contains (data, value) {
    var iterator, iteration;

    if (! assigned(data)) {
      return false;
    ***REMOVED***

    if (haveSets && instanceStrict(data, Set)) {
      return data.has(value);
    ***REMOVED***

    if (string(data)) {
      return data.indexOf(value) !== -1;
    ***REMOVED***

    if (haveSymbols && data[Symbol.iterator] && isFunction(data.values)) {
      iterator = data.values();

      do {
        iteration = iterator.next();

        if (iteration.value === value) {
          return true;
        ***REMOVED***
      ***REMOVED*** while (! iteration.done);

      return false;
    ***REMOVED***

    return some(data, function (key, dataValue) {
      return dataValue === value;
    ***REMOVED***);
  ***REMOVED***

  /**
   * Public function `in`.
   *
   * Returns true if `value` is in `data`, false otherwise.
   * Like `contains`, but with arguments flipped.
   */
  function isIn (value, data) {
    return contains(data, value);
  ***REMOVED***

  /**
   * Public function `containsKey`.
   *
   * Returns true if `data` contains key `key`, false otherwise.
   * Works with objects, arrays and array-likes (including strings).
   */
  function containsKey (data, key) {
    if (! assigned(data)) {
      return false;
    ***REMOVED***

    if (haveMaps && instanceStrict(data, Map)) {
      return data.has(key);
    ***REMOVED***

    if (iterable(data) && ! number(+key)) {
      return false;
    ***REMOVED***

    return !! data[key];
  ***REMOVED***

  /**
   * Public function `keyIn`.
   *
   * Returns true if key `key` is in `data`, false otherwise.
   * Like `contains`, but with arguments flipped.
   */
  function keyIn (key, data) {
    return containsKey(data, key);
  ***REMOVED***

  /**
   * Public function `hasLength`.
   *
   * Returns true if `data` has a length property that equals `length`, false
   * otherwise.
   */
  function hasLength (data, length) {
    return assigned(data) && data.length === length;
  ***REMOVED***

  /**
   * Public function `date`.
   *
   * Returns true if `data` is a valid date, false otherwise.
   */
  function date (data) {
    return instanceStrict(data, Date) && integer(data.getTime());
  ***REMOVED***

  /**
   * Public function `function`.
   *
   * Returns true if `data` is a function, false otherwise.
   */
  function isFunction (data) {
    return typeof data === 'function';
  ***REMOVED***

  /**
   * Public function `throws`.
   *
   * Returns true if `data` is a function that throws, false otherwise.
   */
  function throws (data) {
    if (! isFunction(data)) {
      return false;
    ***REMOVED***

    try {
      data();
    ***REMOVED*** catch (error) {
      return true;
    ***REMOVED***

    return false;
  ***REMOVED***

  /**
   * Public function `map`.
   *
   * Maps each value from `data` to the corresponding predicate and returns
   * the results. If the same function is to be applied across all of the data,
   * a single predicate function may be passed in.
   */
  function map (data, predicates) {
    var result;

    if (isArray(data)) {
      result = [];
    ***REMOVED*** else {
      result = {***REMOVED***;
    ***REMOVED***

    if (isFunction(predicates)) {
      forEach(data, function (key, value) {
        result[key] = predicates(value);
      ***REMOVED***);
    ***REMOVED*** else {
      if (! isArray(predicates)) {
        assert.object(predicates);
      ***REMOVED***

      var dataKeys = keys(data || {***REMOVED***);

      forEach(predicates, function (key, predicate) {
        dataKeys.some(function (dataKey, index) {
          if (dataKey === key) {
            dataKeys.splice(index, 1);
            return true;
          ***REMOVED***
          return false;
        ***REMOVED***);

        if (isFunction(predicate)) {
          if (not.assigned(data)) {
            result[key] = !! predicate.m;
          ***REMOVED*** else {
            result[key] = predicate(data[key]);
          ***REMOVED***
        ***REMOVED*** else {
          result[key] = map(data[key], predicate);
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***

    return result;
  ***REMOVED***

  function forEach (object, action) {
    for (var key in object) {
      if (hasOwnProperty.call(object, key)) {
        action(key, object[key]);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  /**
   * Public function `all`
   *
   * Check that all boolean values are true
   * in an array or object returned from `map`.
   */
  function all (data) {
    if (isArray(data)) {
      return testArray(data, false);
    ***REMOVED***

    assert.object(data);

    return testObject(data, false);
  ***REMOVED***

  function testArray (data, result) {
    var i;

    for (i = 0; i < data.length; i += 1) {
      if (data[i] === result) {
        return result;
      ***REMOVED***
    ***REMOVED***

    return ! result;
  ***REMOVED***

  function testObject (data, result) {
    var key, value;

    for (key in data) {
      if (hasOwnProperty.call(data, key)) {
        value = data[key];

        if (object(value) && testObject(value, result) === result) {
          return result;
        ***REMOVED***

        if (value === result) {
          return result;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    return ! result;
  ***REMOVED***

  /**
   * Public function `any`
   *
   * Check that at least one boolean value is true
   * in an array or object returned from `map`.
   */
  function any (data) {
    if (isArray(data)) {
      return testArray(data, true);
    ***REMOVED***

    assert.object(data);

    return testObject(data, true);
  ***REMOVED***

  function mixin (target, source) {
    forEach(source, function (key, value) {
      target[key] = value;
    ***REMOVED***);

    return target;
  ***REMOVED***

  /**
   * Public modifier `assert`.
   *
   * Throws if `predicate` returns false.
   */
  function assertModifier (predicate, defaultMessage) {
    return function () {
      var args = arguments;
      var argCount = predicate.l || predicate.length;
      var message = args[argCount];
      var ErrorType = args[argCount + 1];

      assertImpl(
        predicate.apply(null, args),
        nonEmptyString(message) ? message : defaultMessage
          .replace('{a***REMOVED***', messageFormatter(args[0]))
          .replace('{e***REMOVED***', messageFormatter(args[1]))
          .replace('{e2***REMOVED***', messageFormatter(args[2]))
          .replace('{t***REMOVED***', function () {
            var arg = args[1];

            if (arg && arg.name) {
              return arg.name;
            ***REMOVED***

            return arg;
          ***REMOVED***),
        isFunction(ErrorType) ? ErrorType : TypeError
      );

      return args[0];
    ***REMOVED***;
  ***REMOVED***

  function messageFormatter (arg) {
    return function () {
      if (string(arg)) {
        return '"' + arg.replace(/\\/g, '\\\\').replace(/"/g, '\\"') + '"';
      ***REMOVED***

      if (arg && arg !== true && arg.constructor && ! instanceStrict(arg, RegExp) && typeof arg !== 'number') {
        return arg.constructor.name;
      ***REMOVED***

      return arg;
    ***REMOVED***;
  ***REMOVED***

  function assertImpl (value, message, ErrorType) {
    if (value) {
      return value;
    ***REMOVED***
    throw new (ErrorType || Error)(message || 'assert failed');
  ***REMOVED***

  /**
   * Public modifier `not`.
   *
   * Negates `predicate`.
   */
  function notModifier (predicate) {
    var modifiedPredicate = function () {
      return notImpl(predicate.apply(null, arguments));
    ***REMOVED***;
    modifiedPredicate.l = predicate.length;
    return modifiedPredicate;
  ***REMOVED***

  function notImpl (value) {
    return ! value;
  ***REMOVED***

  /**
   * Public modifier `maybe`.
   *
   * Returns true if predicate argument is  null or undefined,
   * otherwise propagates the return value from `predicate`.
   */
  function maybeModifier (predicate) {
    var modifiedPredicate = function () {
      if (not.assigned(arguments[0])) {
        return true;
      ***REMOVED***

      return predicate.apply(null, arguments);
    ***REMOVED***;
    modifiedPredicate.l = predicate.length;

    // Hackishly indicate that this is a maybe.xxx predicate.
    // Without this flag, the alternative would be to iterate
    // through the maybe predicates or use indexOf to check,
    // which would be time-consuming.
    modifiedPredicate.m = true;

    return modifiedPredicate;
  ***REMOVED***

  function maybeImpl (value) {
    if (! assigned(value)) {
      return true;
    ***REMOVED***

    return value;
  ***REMOVED***

  /**
   * Public modifier `of`.
   *
   * Applies the chained predicate to members of the collection.
   */
  function ofModifier (target, type, predicate) {
    var modifiedPredicate = function () {
      var collection, args;

      collection = arguments[0];

      if (target === 'maybe' && not.assigned(collection)) {
        return true;
      ***REMOVED***

      if (! type(collection)) {
        return false;
      ***REMOVED***

      collection = coerceCollection(type, collection);
      args = slice.call(arguments, 1);

      try {
        collection.forEach(function (item) {
          if (
            (target !== 'maybe' || assigned(item)) &&
            ! predicate.apply(null, [ item ].concat(args))
          ) {
            // TODO: Replace with for...of when ES6 is required.
            throw 0;
          ***REMOVED***
        ***REMOVED***);
      ***REMOVED*** catch (ignore) {
        return false;
      ***REMOVED***

      return true;
    ***REMOVED***;
    modifiedPredicate.l = predicate.length;
    return modifiedPredicate;
  ***REMOVED***

  function coerceCollection (type, collection) {
    switch (type) {
      case arrayLike:
        return slice.call(collection);
      case object:
        return keys(collection).map(function (key) {
          return collection[key];
        ***REMOVED***);
      default:
        return collection;
    ***REMOVED***
  ***REMOVED***

  function createModifiedPredicates (modifier, object) {
    return createModifiedFunctions([ modifier, predicates, object, '' ]);
  ***REMOVED***

  function createModifiedFunctions (args) {
    var modifier, messageModifier, object, functions;

    modifier = args.shift();
    messageModifier = args.pop();
    object = args.pop();
    functions = args.pop();

    forEach(functions, function (key, fn) {
      var message = messages[key];
      if (message && messageModifier) {
        message = message.replace('to', messageModifier + 'to');
      ***REMOVED***

      Object.defineProperty(object, key, {
        configurable: false,
        enumerable: true,
        writable: false,
        value: modifier.apply(null, args.concat(fn, message))
      ***REMOVED***);
    ***REMOVED***);

    return object;
  ***REMOVED***

  function createModifiedModifier (modifier, modified, messageModifier) {
    return createModifiedFunctions([ modifier, modified, {***REMOVED***, messageModifier ]);
  ***REMOVED***

  function createOfPredicates (key) {
    predicates[key].of = createModifiedFunctions(
      [ ofModifier.bind(null, null), predicates[key], predicates, {***REMOVED***, '' ]
    );
  ***REMOVED***

  function createOfModifiers (base, modifier) {
    collections.forEach(function (key) {
      base[key].of = createModifiedModifier(modifier, predicates[key].of);
    ***REMOVED***);
  ***REMOVED***

  function createMaybeOfModifiers (key) {
    maybe[key].of = createModifiedFunctions(
      [ ofModifier.bind(null, 'maybe'), predicates[key], predicates, {***REMOVED***, '' ]
    );
    assert.maybe[key].of = createModifiedModifier(assertModifier, maybe[key].of);
    assert.not[key].of = createModifiedModifier(assertModifier, not[key].of);
  ***REMOVED***

  function exportFunctions (functions) {
    if (typeof define === 'function' && define.amd) {
      define(function () {
        return functions;
      ***REMOVED***);
    ***REMOVED*** else if (typeof module !== 'undefined' && module !== null && module.exports) {
      module.exports = functions;
    ***REMOVED*** else {
      globals.check = functions;
    ***REMOVED***
  ***REMOVED***
***REMOVED***(this));
