// ajv/lib/vocabularies/validation/required.ts
import type {CodeKeywordDefinition, ErrorObject, KeywordErrorDefinition***REMOVED*** from "../../types"
import type {KeywordCxt***REMOVED*** from "../../compile/validate"
import {
  checkReportMissingProp,
  checkMissingProp,
  reportMissingProp,
  propertyInData,
  noPropertyInData,
***REMOVED*** from "../code"
import {_, str, nil, not, Name, Code***REMOVED*** from "../../compile/codegen"
import {checkStrictMode***REMOVED*** from "../../compile/util"

export type RequiredError = ErrorObject<
  "required",
  {missingProperty: string***REMOVED***,
  string[] | {$data: string***REMOVED***
>

const error: KeywordErrorDefinition = {
  message: ({params: {missingProperty***REMOVED******REMOVED***) => str`must have required property '${missingProperty***REMOVED***'`,
  params: ({params: {missingProperty***REMOVED******REMOVED***) => _`{missingProperty: ${missingProperty***REMOVED******REMOVED***`,
***REMOVED***

const def: CodeKeywordDefinition = {
  keyword: "required",
  type: "object",
  schemaType: "array",
  $data: true,
  error,
  code(cxt: KeywordCxt) {
    const {gen, schema, schemaCode, data, $data, it***REMOVED*** = cxt
    const {opts***REMOVED*** = it
    if (!$data && schema.length === 0) return
    const useLoop = schema.length >= opts.loopRequired
    if (it.allErrors) allErrorsMode()
    else exitOnErrorMode()

    if (opts.strictRequired) {
      const props = cxt.parentSchema.properties
      const {definedProperties***REMOVED*** = cxt.it
      for (const requiredKey of schema) {
        if (props?.[requiredKey] === undefined && !definedProperties.has(requiredKey)) {
          const schemaPath = it.schemaEnv.baseId + it.errSchemaPath
          const msg = `required property "${requiredKey***REMOVED***" is not defined at "${schemaPath***REMOVED***" (strictRequired)`
          checkStrictMode(it, msg, it.opts.strictRequired)
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    function allErrorsMode(): void {
      if (useLoop || $data) {
        cxt.block$data(nil, loopAllRequired)
      ***REMOVED*** else {
        for (const prop of schema) {
          checkReportMissingProp(cxt, prop)
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    function exitOnErrorMode(): void {
      const missing = gen.let("missing")
      if (useLoop || $data) {
        const valid = gen.let("valid", true)
        cxt.block$data(valid, () => loopUntilMissing(missing, valid))
        cxt.ok(valid)
      ***REMOVED*** else {
        gen.if(checkMissingProp(cxt, schema, missing))
        reportMissingProp(cxt, missing)
        gen.else()
      ***REMOVED***
    ***REMOVED***

    function loopAllRequired(): void {
      gen.forOf("prop", schemaCode as Code, (prop) => {
        cxt.setParams({missingProperty: prop***REMOVED***)
        gen.if(noPropertyInData(gen, data, prop, opts.ownProperties), () => cxt.error())
      ***REMOVED***)
    ***REMOVED***

    function loopUntilMissing(missing: Name, valid: Name): void {
      cxt.setParams({missingProperty: missing***REMOVED***)
      gen.forOf(
        missing,
        schemaCode as Code,
        () => {
          gen.assign(valid, propertyInData(gen, data, missing, opts.ownProperties))
          gen.if(not(valid), () => {
            cxt.error()
            gen.break()
          ***REMOVED***)
        ***REMOVED***,
        nil
      )
    ***REMOVED***
  ***REMOVED***,
***REMOVED***

export default def
