// @babel/traverse/lib/scope/lib/renamer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var t = require("@babel/types");
var _t = t;
var _traverseNode = require("../../traverse-node.js");
var _visitors = require("../../visitors.js");
var _context = require("../../path/context.js");
const {
  getAssignmentIdentifiers
***REMOVED*** = _t;
const renameVisitor = {
  ReferencedIdentifier({
    node
  ***REMOVED***, state) {
    if (node.name === state.oldName) {
      node.name = state.newName;
    ***REMOVED***
  ***REMOVED***,
  Scope(path, state) {
    if (!path.scope.bindingIdentifierEquals(state.oldName, state.binding.identifier)) {
      path.skip();
      if (path.isMethod()) {
        if (!path.requeueComputedKeyAndDecorators) {
          _context.requeueComputedKeyAndDecorators.call(path);
        ***REMOVED*** else {
          path.requeueComputedKeyAndDecorators();
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***,
  ObjectProperty({
    node,
    scope
  ***REMOVED***, state) {
    const {
      name
    ***REMOVED*** = node.key;
    if (node.shorthand && (name === state.oldName || name === state.newName) && scope.getBindingIdentifier(name) === state.binding.identifier) {
      node.shorthand = false;
      {
        var _node$extra;
        if ((_node$extra = node.extra) != null && _node$extra.shorthand) node.extra.shorthand = false;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***,
  "AssignmentExpression|Declaration|VariableDeclarator"(path, state) {
    if (path.isVariableDeclaration()) return;
    const ids = path.isAssignmentExpression() ? getAssignmentIdentifiers(path.node) : path.getOuterBindingIdentifiers();
    for (const name in ids) {
      if (name === state.oldName) ids[name].name = state.newName;
    ***REMOVED***
  ***REMOVED***
***REMOVED***;
class Renamer {
  constructor(binding, oldName, newName) {
    this.newName = newName;
    this.oldName = oldName;
    this.binding = binding;
  ***REMOVED***
  maybeConvertFromExportDeclaration(parentDeclar) {
    const maybeExportDeclar = parentDeclar.parentPath;
    if (!maybeExportDeclar.isExportDeclaration()) {
      return;
    ***REMOVED***
    if (maybeExportDeclar.isExportDefaultDeclaration()) {
      const {
        declaration
      ***REMOVED*** = maybeExportDeclar.node;
      if (t.isDeclaration(declaration) && !declaration.id) {
        return;
      ***REMOVED***
    ***REMOVED***
    if (maybeExportDeclar.isExportAllDeclaration()) {
      return;
    ***REMOVED***
    maybeExportDeclar.splitExportDeclaration();
  ***REMOVED***
  maybeConvertFromClassFunctionDeclaration(path) {
    return path;
  ***REMOVED***
  maybeConvertFromClassFunctionExpression(path) {
    return path;
  ***REMOVED***
  rename() {
    const {
      binding,
      oldName,
      newName
    ***REMOVED*** = this;
    const {
      scope,
      path
    ***REMOVED*** = binding;
    const parentDeclar = path.find(path => path.isDeclaration() || path.isFunctionExpression() || path.isClassExpression());
    if (parentDeclar) {
      const bindingIds = parentDeclar.getOuterBindingIdentifiers();
      if (bindingIds[oldName] === binding.identifier) {
        this.maybeConvertFromExportDeclaration(parentDeclar);
      ***REMOVED***
    ***REMOVED***
    const blockToTraverse = arguments[0] || scope.block;
    const skipKeys = {
      discriminant: true
    ***REMOVED***;
    if (t.isMethod(blockToTraverse)) {
      if (blockToTraverse.computed) {
        skipKeys.key = true;
      ***REMOVED***
      if (!t.isObjectMethod(blockToTraverse)) {
        skipKeys.decorators = true;
      ***REMOVED***
    ***REMOVED***
    (0, _traverseNode.traverseNode)(blockToTraverse, (0, _visitors.explode)(renameVisitor), scope, this, scope.path, skipKeys);
    if (!arguments[0]) {
      scope.removeOwnBinding(oldName);
      scope.bindings[newName] = binding;
      this.binding.identifier.name = newName;
    ***REMOVED***
    if (parentDeclar) {
      this.maybeConvertFromClassFunctionDeclaration(path);
      this.maybeConvertFromClassFunctionExpression(path);
    ***REMOVED***
  ***REMOVED***
***REMOVED***
exports.default = Renamer;
