// @babel/helpers/lib/helpers/asyncIterator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = _asyncIterator;
function _asyncIterator(iterable) {
  var method,
    async,
    sync,
    retry = 2;
  if (typeof Symbol !== "undefined") {
    async = Symbol.asyncIterator;
    sync = Symbol.iterator;
  ***REMOVED***
  while (retry--) {
    if (async && (method = iterable[async]) != null) {
      return method.call(iterable);
    ***REMOVED***
    if (sync && (method = iterable[sync]) != null) {
      return new AsyncFromSyncIterator(method.call(iterable));
    ***REMOVED***
    async = "@@asyncIterator";
    sync = "@@iterator";
  ***REMOVED***
  throw new TypeError("Object is not async iterable");
***REMOVED***
function AsyncFromSyncIterator(s) {
  AsyncFromSyncIterator = function (s) {
    this.s = s;
    this.n = s.next;
  ***REMOVED***;
  AsyncFromSyncIterator.prototype = {
    s: null,
    n: null,
    next: function () {
      return AsyncFromSyncIteratorContinuation(this.n.apply(this.s, arguments));
    ***REMOVED***,
    return: function (value) {
      var ret = this.s.return;
      if (ret === undefined) {
        return Promise.resolve({
          value: value,
          done: true
        ***REMOVED***);
      ***REMOVED***
      return AsyncFromSyncIteratorContinuation(ret.apply(this.s, arguments));
    ***REMOVED***,
    throw: function (maybeError) {
      var thr = this.s.return;
      if (thr === undefined) {
        return Promise.reject(maybeError);
      ***REMOVED***
      return AsyncFromSyncIteratorContinuation(thr.apply(this.s, arguments));
    ***REMOVED***
  ***REMOVED***;
  function AsyncFromSyncIteratorContinuation(r) {
    if (Object(r) !== r) {
      return Promise.reject(new TypeError(r + " is not an object."));
    ***REMOVED***
    var done = r.done;
    return Promise.resolve(r.value).then(function (value) {
      return {
        value: value,
        done: done
      ***REMOVED***;
    ***REMOVED***);
  ***REMOVED***
  return new AsyncFromSyncIterator(s);
***REMOVED***
