// @babel/helpers/lib/helpers/dispose.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = _dispose;
function dispose_SuppressedError(error, suppressed) {
  if (typeof SuppressedError !== "undefined") {
    dispose_SuppressedError = SuppressedError;
  ***REMOVED*** else {
    dispose_SuppressedError = function SuppressedError(error, suppressed) {
      this.suppressed = suppressed;
      this.error = error;
      this.stack = new Error().stack;
    ***REMOVED***;
    dispose_SuppressedError.prototype = Object.create(Error.prototype, {
      constructor: {
        value: dispose_SuppressedError,
        writable: true,
        configurable: true
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***
  return new dispose_SuppressedError(error, suppressed);
***REMOVED***
function _dispose(stack, error, hasError) {
  function next() {
    while (stack.length > 0) {
      try {
        var r = stack.pop();
        var p = r.d.call(r.v);
        if (r.a) return Promise.resolve(p).then(next, err);
      ***REMOVED*** catch (e) {
        return err(e);
      ***REMOVED***
    ***REMOVED***
    if (hasError) throw error;
  ***REMOVED***
  function err(e) {
    error = hasError ? new dispose_SuppressedError(error, e) : e;
    hasError = true;
    return next();
  ***REMOVED***
  return next();
***REMOVED***
