// @babel/helpers/lib/helpers/regeneratorRuntime.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = _regeneratorRuntime;
function _regeneratorRuntime() {
  "use strict";
  exports.default = _regeneratorRuntime = function () {
    return _exports;
  ***REMOVED***;
  var _exports = {***REMOVED***;
  var Op = Object.prototype;
  var hasOwn = Op.hasOwnProperty;
  var undefined;
  var $Symbol = typeof Symbol === "function" ? Symbol : {***REMOVED***;
  var iteratorSymbol = $Symbol.iterator || "@@iterator";
  var asyncIteratorSymbol = $Symbol.asyncIterator || "@@asyncIterator";
  var toStringTagSymbol = $Symbol.toStringTag || "@@toStringTag";
  function define(obj, key, value, noFlags) {
    return Object.defineProperty(obj, key, {
      value: value,
      enumerable: !noFlags,
      configurable: !noFlags,
      writable: !noFlags
    ***REMOVED***);
  ***REMOVED***
  try {
    define({***REMOVED***, "");
  ***REMOVED*** catch (_) {
    define = function (obj, key, value) {
      return obj[key] = value;
    ***REMOVED***;
  ***REMOVED***
  function wrap(innerFn, outerFn, self, tryLocsList) {
    var protoGenerator = outerFn && outerFn.prototype instanceof Generator ? outerFn : Generator;
    var generator = Object.create(protoGenerator.prototype);
    var context = new Context(tryLocsList || []);
    define(generator, "_invoke", makeInvokeMethod(innerFn, self, context), true);
    return generator;
  ***REMOVED***
  _exports.wrap = wrap;
  function tryCatch(fn, obj, arg) {
    try {
      return {
        type: "normal",
        arg: fn.call(obj, arg)
      ***REMOVED***;
    ***REMOVED*** catch (err) {
      return {
        type: "throw",
        arg: err
      ***REMOVED***;
    ***REMOVED***
  ***REMOVED***
  var ContinueSentinel = {***REMOVED***;
  function Generator() {***REMOVED***
  function GeneratorFunction() {***REMOVED***
  function GeneratorFunctionPrototype() {***REMOVED***
  var IteratorPrototype = {***REMOVED***;
  define(IteratorPrototype, iteratorSymbol, function () {
    return this;
  ***REMOVED***);
  var getProto = Object.getPrototypeOf;
  var NativeIteratorPrototype = getProto && getProto(getProto(values([])));
  if (NativeIteratorPrototype && NativeIteratorPrototype !== Op && hasOwn.call(NativeIteratorPrototype, iteratorSymbol)) {
    IteratorPrototype = NativeIteratorPrototype;
  ***REMOVED***
  var Gp = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(IteratorPrototype);
  GeneratorFunction.prototype = GeneratorFunctionPrototype;
  define(Gp, "constructor", GeneratorFunctionPrototype);
  define(GeneratorFunctionPrototype, "constructor", GeneratorFunction);
  GeneratorFunction.displayName = define(GeneratorFunctionPrototype, toStringTagSymbol, "GeneratorFunction");
  function defineIteratorMethods(prototype) {
    ["next", "throw", "return"].forEach(function (method) {
      define(prototype, method, function (arg) {
        return this._invoke(method, arg);
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***
  _exports.isGeneratorFunction = function (genFun) {
    var ctor = typeof genFun === "function" && genFun.constructor;
    return ctor ? ctor === GeneratorFunction || (ctor.displayName || ctor.name) === "GeneratorFunction" : false;
  ***REMOVED***;
  _exports.mark = function (genFun) {
    if (Object.setPrototypeOf) {
      Object.setPrototypeOf(genFun, GeneratorFunctionPrototype);
    ***REMOVED*** else {
      genFun.__proto__ = GeneratorFunctionPrototype;
      define(genFun, toStringTagSymbol, "GeneratorFunction");
    ***REMOVED***
    genFun.prototype = Object.create(Gp);
    return genFun;
  ***REMOVED***;
  _exports.awrap = function (arg) {
    return {
      __await: arg
    ***REMOVED***;
  ***REMOVED***;
  function AsyncIterator(generator, PromiseImpl) {
    function invoke(method, arg, resolve, reject) {
      var record = tryCatch(generator[method], generator, arg);
      if (record.type === "throw") {
        reject(record.arg);
      ***REMOVED*** else {
        var result = record.arg;
        var value = result.value;
        if (value && typeof value === "object" && hasOwn.call(value, "__await")) {
          return PromiseImpl.resolve(value.__await).then(function (value) {
            invoke("next", value, resolve, reject);
          ***REMOVED***, function (err) {
            invoke("throw", err, resolve, reject);
          ***REMOVED***);
        ***REMOVED***
        return PromiseImpl.resolve(value).then(function (unwrapped) {
          result.value = unwrapped;
          resolve(result);
        ***REMOVED***, function (error) {
          return invoke("throw", error, resolve, reject);
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***
    var previousPromise;
    function enqueue(method, arg) {
      function callInvokeWithMethodAndArg() {
        return new PromiseImpl(function (resolve, reject) {
          invoke(method, arg, resolve, reject);
        ***REMOVED***);
      ***REMOVED***
      return previousPromise = previousPromise ? previousPromise.then(callInvokeWithMethodAndArg, callInvokeWithMethodAndArg) : callInvokeWithMethodAndArg();
    ***REMOVED***
    define(this, "_invoke", enqueue, true);
  ***REMOVED***
  defineIteratorMethods(AsyncIterator.prototype);
  define(AsyncIterator.prototype, asyncIteratorSymbol, function () {
    return this;
  ***REMOVED***);
  _exports.AsyncIterator = AsyncIterator;
  _exports.async = function (innerFn, outerFn, self, tryLocsList, PromiseImpl) {
    if (PromiseImpl === void 0) PromiseImpl = Promise;
    var iter = new AsyncIterator(wrap(innerFn, outerFn, self, tryLocsList), PromiseImpl);
    return _exports.isGeneratorFunction(outerFn) ? iter : iter.next().then(function (result) {
      return result.done ? result.value : iter.next();
    ***REMOVED***);
  ***REMOVED***;
  function makeInvokeMethod(innerFn, self, context) {
    var state = 1;
    return function invoke(method, arg) {
      if (state === 3) {
        throw new Error("Generator is already running");
      ***REMOVED***
      if (state === 4) {
        if (method === "throw") {
          throw arg;
        ***REMOVED***
        return {
          value: undefined,
          done: true
        ***REMOVED***;
      ***REMOVED***
      context.method = method;
      context.arg = arg;
      while (true) {
        var delegate = context.delegate;
        if (delegate) {
          var delegateResult = maybeInvokeDelegate(delegate, context);
          if (delegateResult) {
            if (delegateResult === ContinueSentinel) continue;
            return delegateResult;
          ***REMOVED***
        ***REMOVED***
        if (context.method === "next") {
          context.sent = context._sent = context.arg;
        ***REMOVED*** else if (context.method === "throw") {
          if (state === 1) {
            state = 4;
            throw context.arg;
          ***REMOVED***
          context.dispatchException(context.arg);
        ***REMOVED*** else if (context.method === "return") {
          context.abrupt("return", context.arg);
        ***REMOVED***
        state = 3;
        var record = tryCatch(innerFn, self, context);
        if (record.type === "normal") {
          state = context.done ? 4 : 2;
          if (record.arg === ContinueSentinel) {
            continue;
          ***REMOVED***
          return {
            value: record.arg,
            done: context.done
          ***REMOVED***;
        ***REMOVED*** else if (record.type === "throw") {
          state = 4;
          context.method = "throw";
          context.arg = record.arg;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***
  function maybeInvokeDelegate(delegate, context) {
    var methodName = context.method;
    var method = delegate.i[methodName];
    if (method === undefined) {
      context.delegate = null;
      if (methodName === "throw" && delegate.i["return"]) {
        context.method = "return";
        context.arg = undefined;
        maybeInvokeDelegate(delegate, context);
        if (context.method === "throw") {
          return ContinueSentinel;
        ***REMOVED***
      ***REMOVED***
      if (methodName !== "return") {
        context.method = "throw";
        context.arg = new TypeError("The iterator does not provide a '" + methodName + "' method");
      ***REMOVED***
      return ContinueSentinel;
    ***REMOVED***
    var record = tryCatch(method, delegate.i, context.arg);
    if (record.type === "throw") {
      context.method = "throw";
      context.arg = record.arg;
      context.delegate = null;
      return ContinueSentinel;
    ***REMOVED***
    var info = record.arg;
    if (!info) {
      context.method = "throw";
      context.arg = new TypeError("iterator result is not an object");
      context.delegate = null;
      return ContinueSentinel;
    ***REMOVED***
    if (info.done) {
      context[delegate.r] = info.value;
      context.next = delegate.n;
      if (context.method !== "return") {
        context.method = "next";
        context.arg = undefined;
      ***REMOVED***
    ***REMOVED*** else {
      return info;
    ***REMOVED***
    context.delegate = null;
    return ContinueSentinel;
  ***REMOVED***
  defineIteratorMethods(Gp);
  define(Gp, toStringTagSymbol, "Generator");
  define(Gp, iteratorSymbol, function () {
    return this;
  ***REMOVED***);
  define(Gp, "toString", function () {
    return "[object Generator]";
  ***REMOVED***);
  function pushTryEntry(locs) {
    this.tryEntries.push(locs);
  ***REMOVED***
  function resetTryEntry(entry) {
    var record = entry[4] || {***REMOVED***;
    record.type = "normal";
    record.arg = undefined;
    entry[4] = record;
  ***REMOVED***
  function Context(tryLocsList) {
    this.tryEntries = [[-1]];
    tryLocsList.forEach(pushTryEntry, this);
    this.reset(true);
  ***REMOVED***
  _exports.keys = function (val) {
    var object = Object(val);
    var keys = [];
    var key;
    for (var key in object) {
      keys.unshift(key);
    ***REMOVED***
    return function next() {
      while (keys.length) {
        key = keys.pop();
        if (key in object) {
          next.value = key;
          next.done = false;
          return next;
        ***REMOVED***
      ***REMOVED***
      next.done = true;
      return next;
    ***REMOVED***;
  ***REMOVED***;
  function values(iterable) {
    if (iterable != null) {
      var iteratorMethod = iterable[iteratorSymbol];
      if (iteratorMethod) {
        return iteratorMethod.call(iterable);
      ***REMOVED***
      if (typeof iterable.next === "function") {
        return iterable;
      ***REMOVED***
      if (!isNaN(iterable.length)) {
        var i = -1,
          next = function next() {
            while (++i < iterable.length) {
              if (hasOwn.call(iterable, i)) {
                next.value = iterable[i];
                next.done = false;
                return next;
              ***REMOVED***
            ***REMOVED***
            next.value = undefined;
            next.done = true;
            return next;
          ***REMOVED***;
        return next.next = next;
      ***REMOVED***
    ***REMOVED***
    throw new TypeError(typeof iterable + " is not iterable");
  ***REMOVED***
  _exports.values = values;
  Context.prototype = {
    constructor: Context,
    reset: function (skipTempReset) {
      this.prev = this.next = 0;
      this.sent = this._sent = undefined;
      this.done = false;
      this.delegate = null;
      this.method = "next";
      this.arg = undefined;
      this.tryEntries.forEach(resetTryEntry);
      if (!skipTempReset) {
        for (var name in this) {
          if (name.charAt(0) === "t" && hasOwn.call(this, name) && !isNaN(+name.slice(1))) {
            this[name] = undefined;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***,
    stop: function () {
      this.done = true;
      var rootEntry = this.tryEntries[0];
      var rootRecord = rootEntry[4];
      if (rootRecord.type === "throw") {
        throw rootRecord.arg;
      ***REMOVED***
      return this.rval;
    ***REMOVED***,
    dispatchException: function (exception) {
      if (this.done) {
        throw exception;
      ***REMOVED***
      var context = this;
      function handle(loc) {
        record.type = "throw";
        record.arg = exception;
        context.next = loc;
      ***REMOVED***
      for (var i = context.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        var record = entry[4];
        var prev = this.prev;
        var catchLoc = entry[1];
        var finallyLoc = entry[2];
        if (entry[0] === -1) {
          handle("end");
          return false;
        ***REMOVED***
        if (!catchLoc && !finallyLoc) {
          throw new Error("try statement without catch or finally");
        ***REMOVED***
        if (entry[0] != null && entry[0] <= prev) {
          if (prev < catchLoc) {
            this.method = "next";
            this.arg = undefined;
            handle(catchLoc);
            return true;
          ***REMOVED*** else if (prev < finallyLoc) {
            handle(finallyLoc);
            return false;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***,
    abrupt: function (type, arg) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        if (entry[0] > -1 && entry[0] <= this.prev && this.prev < entry[2]) {
          var finallyEntry = entry;
          break;
        ***REMOVED***
      ***REMOVED***
      if (finallyEntry && (type === "break" || type === "continue") && finallyEntry[0] <= arg && arg <= finallyEntry[2]) {
        finallyEntry = null;
      ***REMOVED***
      var record = finallyEntry ? finallyEntry[4] : {***REMOVED***;
      record.type = type;
      record.arg = arg;
      if (finallyEntry) {
        this.method = "next";
        this.next = finallyEntry[2];
        return ContinueSentinel;
      ***REMOVED***
      return this.complete(record);
    ***REMOVED***,
    complete: function (record, afterLoc) {
      if (record.type === "throw") {
        throw record.arg;
      ***REMOVED***
      if (record.type === "break" || record.type === "continue") {
        this.next = record.arg;
      ***REMOVED*** else if (record.type === "return") {
        this.rval = this.arg = record.arg;
        this.method = "return";
        this.next = "end";
      ***REMOVED*** else if (record.type === "normal" && afterLoc) {
        this.next = afterLoc;
      ***REMOVED***
      return ContinueSentinel;
    ***REMOVED***,
    finish: function (finallyLoc) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        if (entry[2] === finallyLoc) {
          this.complete(entry[4], entry[3]);
          resetTryEntry(entry);
          return ContinueSentinel;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***,
    catch: function (tryLoc) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        if (entry[0] === tryLoc) {
          var record = entry[4];
          if (record.type === "throw") {
            var thrown = record.arg;
            resetTryEntry(entry);
          ***REMOVED***
          return thrown;
        ***REMOVED***
      ***REMOVED***
      throw new Error("illegal catch attempt");
    ***REMOVED***,
    delegateYield: function (iterable, resultName, nextLoc) {
      this.delegate = {
        i: values(iterable),
        r: resultName,
        n: nextLoc
      ***REMOVED***;
      if (this.method === "next") {
        this.arg = undefined;
      ***REMOVED***
      return ContinueSentinel;
    ***REMOVED***
  ***REMOVED***;
  return _exports;
***REMOVED***
