// @typescript-eslint/eslint-plugin/dist/rules/naming-convention-utils/format.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.PredefinedFormatToCheckFunction = void 0;
const enums_1 = require("./enums");
/*
These format functions are taken from `tslint-consistent-codestyle/naming-convention`:
https://github.com/ajafff/tslint-consistent-codestyle/blob/ab156cc8881bcc401236d999f4ce034b59039e81/rules/namingConventionRule.ts#L603-L645

The license for the code can be viewed here:
https://github.com/ajafff/tslint-consistent-codestyle/blob/ab156cc8881bcc401236d999f4ce034b59039e81/LICENSE
*/
/*
Why not regex here? Because it's actually really, really difficult to create a regex to handle
all of the unicode cases, and we have many non-english users that use non-english characters.
https://gist.github.com/mathiasbynens/6334847
*/
function isPascalCase(name) {
    return (name.length === 0 ||
        (name[0] === name[0].toUpperCase() && !name.includes('_')));
***REMOVED***
function isStrictPascalCase(name) {
    return (name.length === 0 ||
        (name[0] === name[0].toUpperCase() && hasStrictCamelHumps(name, true)));
***REMOVED***
function isCamelCase(name) {
    return (name.length === 0 ||
        (name[0] === name[0].toLowerCase() && !name.includes('_')));
***REMOVED***
function isStrictCamelCase(name) {
    return (name.length === 0 ||
        (name[0] === name[0].toLowerCase() && hasStrictCamelHumps(name, false)));
***REMOVED***
function hasStrictCamelHumps(name, isUpper) {
    function isUppercaseChar(char) {
        return char === char.toUpperCase() && char !== char.toLowerCase();
    ***REMOVED***
    if (name.startsWith('_')) {
        return false;
    ***REMOVED***
    for (let i = 1; i < name.length; ++i) {
        if (name[i] === '_') {
            return false;
        ***REMOVED***
        if (isUpper === isUppercaseChar(name[i])) {
            if (isUpper) {
                return false;
            ***REMOVED***
        ***REMOVED***
        else {
            isUpper = !isUpper;
        ***REMOVED***
    ***REMOVED***
    return true;
***REMOVED***
function isSnakeCase(name) {
    return (name.length === 0 ||
        (name === name.toLowerCase() && validateUnderscores(name)));
***REMOVED***
function isUpperCase(name) {
    return (name.length === 0 ||
        (name === name.toUpperCase() && validateUnderscores(name)));
***REMOVED***
/** Check for leading trailing and adjacent underscores */
function validateUnderscores(name) {
    if (name.startsWith('_')) {
        return false;
    ***REMOVED***
    let wasUnderscore = false;
    for (let i = 1; i < name.length; ++i) {
        if (name[i] === '_') {
            if (wasUnderscore) {
                return false;
            ***REMOVED***
            wasUnderscore = true;
        ***REMOVED***
        else {
            wasUnderscore = false;
        ***REMOVED***
    ***REMOVED***
    return !wasUnderscore;
***REMOVED***
const PredefinedFormatToCheckFunction = {
    [enums_1.PredefinedFormats.PascalCase]: isPascalCase,
    [enums_1.PredefinedFormats.StrictPascalCase]: isStrictPascalCase,
    [enums_1.PredefinedFormats.camelCase]: isCamelCase,
    [enums_1.PredefinedFormats.strictCamelCase]: isStrictCamelCase,
    [enums_1.PredefinedFormats.UPPER_CASE]: isUpperCase,
    [enums_1.PredefinedFormats.snake_case]: isSnakeCase,
***REMOVED***;
exports.PredefinedFormatToCheckFunction = PredefinedFormatToCheckFunction;
