// gensync/index.js
"use strict";

// These use the global symbol registry so that multiple copies of this
// library can work together in case they are not deduped.
const GENSYNC_START = Symbol.for("gensync:v1:start");
const GENSYNC_SUSPEND = Symbol.for("gensync:v1:suspend");

const GENSYNC_EXPECTED_START = "GENSYNC_EXPECTED_START";
const GENSYNC_EXPECTED_SUSPEND = "GENSYNC_EXPECTED_SUSPEND";
const GENSYNC_OPTIONS_ERROR = "GENSYNC_OPTIONS_ERROR";
const GENSYNC_RACE_NONEMPTY = "GENSYNC_RACE_NONEMPTY";
const GENSYNC_ERRBACK_NO_CALLBACK = "GENSYNC_ERRBACK_NO_CALLBACK";

module.exports = Object.assign(
  function gensync(optsOrFn) {
    let genFn = optsOrFn;
    if (typeof optsOrFn !== "function") {
      genFn = newGenerator(optsOrFn);
    ***REMOVED*** else {
      genFn = wrapGenerator(optsOrFn);
    ***REMOVED***

    return Object.assign(genFn, makeFunctionAPI(genFn));
  ***REMOVED***,
  {
    all: buildOperation({
      name: "all",
      arity: 1,
      sync: function(args) {
        const items = Array.from(args[0]);
        return items.map(item => evaluateSync(item));
      ***REMOVED***,
      async: function(args, resolve, reject) {
        const items = Array.from(args[0]);

        if (items.length === 0) {
          Promise.resolve().then(() => resolve([]));
          return;
        ***REMOVED***

        let count = 0;
        const results = items.map(() => undefined);
        items.forEach((item, i) => {
          evaluateAsync(
            item,
            val => {
              results[i] = val;
              count += 1;

              if (count === results.length) resolve(results);
            ***REMOVED***,
            reject
          );
        ***REMOVED***);
      ***REMOVED***,
    ***REMOVED***),
    race: buildOperation({
      name: "race",
      arity: 1,
      sync: function(args) {
        const items = Array.from(args[0]);
        if (items.length === 0) {
          throw makeError("Must race at least 1 item", GENSYNC_RACE_NONEMPTY);
        ***REMOVED***

        return evaluateSync(items[0]);
      ***REMOVED***,
      async: function(args, resolve, reject) {
        const items = Array.from(args[0]);
        if (items.length === 0) {
          throw makeError("Must race at least 1 item", GENSYNC_RACE_NONEMPTY);
        ***REMOVED***

        for (const item of items) {
          evaluateAsync(item, resolve, reject);
        ***REMOVED***
      ***REMOVED***,
    ***REMOVED***),
  ***REMOVED***
);

/**
 * Given a generator function, return the standard API object that executes
 * the generator and calls the callbacks.
 */
function makeFunctionAPI(genFn) {
  const fns = {
    sync: function(...args) {
      return evaluateSync(genFn.apply(this, args));
    ***REMOVED***,
    async: function(...args) {
      return new Promise((resolve, reject) => {
        evaluateAsync(genFn.apply(this, args), resolve, reject);
      ***REMOVED***);
    ***REMOVED***,
    errback: function(...args) {
      const cb = args.pop();
      if (typeof cb !== "function") {
        throw makeError(
          "Asynchronous function called without callback",
          GENSYNC_ERRBACK_NO_CALLBACK
        );
      ***REMOVED***

      let gen;
      try {
        gen = genFn.apply(this, args);
      ***REMOVED*** catch (err) {
        cb(err);
        return;
      ***REMOVED***

      evaluateAsync(gen, val => cb(undefined, val), err => cb(err));
    ***REMOVED***,
  ***REMOVED***;
  return fns;
***REMOVED***

function assertTypeof(type, name, value, allowUndefined) {
  if (
    typeof value === type ||
    (allowUndefined && typeof value === "undefined")
  ) {
    return;
  ***REMOVED***

  let msg;
  if (allowUndefined) {
    msg = `Expected opts.${name***REMOVED*** to be either a ${type***REMOVED***, or undefined.`;
  ***REMOVED*** else {
    msg = `Expected opts.${name***REMOVED*** to be a ${type***REMOVED***.`;
  ***REMOVED***

  throw makeError(msg, GENSYNC_OPTIONS_ERROR);
***REMOVED***
function makeError(msg, code) {
  return Object.assign(new Error(msg), { code ***REMOVED***);
***REMOVED***

/**
 * Given an options object, return a new generator that dispatches the
 * correct handler based on sync or async execution.
 */
function newGenerator({ name, arity, sync, async, errback ***REMOVED***) {
  assertTypeof("string", "name", name, true /* allowUndefined */);
  assertTypeof("number", "arity", arity, true /* allowUndefined */);
  assertTypeof("function", "sync", sync);
  assertTypeof("function", "async", async, true /* allowUndefined */);
  assertTypeof("function", "errback", errback, true /* allowUndefined */);
  if (async && errback) {
    throw makeError(
      "Expected one of either opts.async or opts.errback, but got _both_.",
      GENSYNC_OPTIONS_ERROR
    );
  ***REMOVED***

  if (typeof name !== "string") {
    let fnName;
    if (errback && errback.name && errback.name !== "errback") {
      fnName = errback.name;
    ***REMOVED***
    if (async && async.name && async.name !== "async") {
      fnName = async.name.replace(/Async$/, "");
    ***REMOVED***
    if (sync && sync.name && sync.name !== "sync") {
      fnName = sync.name.replace(/Sync$/, "");
    ***REMOVED***

    if (typeof fnName === "string") {
      name = fnName;
    ***REMOVED***
  ***REMOVED***

  if (typeof arity !== "number") {
    arity = sync.length;
  ***REMOVED***

  return buildOperation({
    name,
    arity,
    sync: function(args) {
      return sync.apply(this, args);
    ***REMOVED***,
    async: function(args, resolve, reject) {
      if (async) {
        async.apply(this, args).then(resolve, reject);
      ***REMOVED*** else if (errback) {
        errback.call(this, ...args, (err, value) => {
          if (err == null) resolve(value);
          else reject(err);
        ***REMOVED***);
      ***REMOVED*** else {
        resolve(sync.apply(this, args));
      ***REMOVED***
    ***REMOVED***,
  ***REMOVED***);
***REMOVED***

function wrapGenerator(genFn) {
  return setFunctionMetadata(genFn.name, genFn.length, function(...args) {
    return genFn.apply(this, args);
  ***REMOVED***);
***REMOVED***

function buildOperation({ name, arity, sync, async ***REMOVED***) {
  return setFunctionMetadata(name, arity, function*(...args) {
    const resume = yield GENSYNC_START;
    if (!resume) {
      // Break the tail call to avoid a bug in V8 v6.X with --harmony enabled.
      const res = sync.call(this, args);
      return res;
    ***REMOVED***

    let result;
    try {
      async.call(
        this,
        args,
        value => {
          if (result) return;

          result = { value ***REMOVED***;
          resume();
        ***REMOVED***,
        err => {
          if (result) return;

          result = { err ***REMOVED***;
          resume();
        ***REMOVED***
      );
    ***REMOVED*** catch (err) {
      result = { err ***REMOVED***;
      resume();
    ***REMOVED***

    // Suspend until the callbacks run. Will resume synchronously if the
    // callback was already called.
    yield GENSYNC_SUSPEND;

    if (result.hasOwnProperty("err")) {
      throw result.err;
    ***REMOVED***

    return result.value;
  ***REMOVED***);
***REMOVED***

function evaluateSync(gen) {
  let value;
  while (!({ value ***REMOVED*** = gen.next()).done) {
    assertStart(value, gen);
  ***REMOVED***
  return value;
***REMOVED***

function evaluateAsync(gen, resolve, reject) {
  (function step() {
    try {
      let value;
      while (!({ value ***REMOVED*** = gen.next()).done) {
        assertStart(value, gen);

        // If this throws, it is considered to have broken the contract
        // established for async handlers. If these handlers are called
        // synchronously, it is also considered bad behavior.
        let sync = true;
        let didSyncResume = false;
        const out = gen.next(() => {
          if (sync) {
            didSyncResume = true;
          ***REMOVED*** else {
            step();
          ***REMOVED***
        ***REMOVED***);
        sync = false;

        assertSuspend(out, gen);

        if (!didSyncResume) {
          // Callback wasn't called synchronously, so break out of the loop
          // and let it call 'step' later.
          return;
        ***REMOVED***
      ***REMOVED***

      return resolve(value);
    ***REMOVED*** catch (err) {
      return reject(err);
    ***REMOVED***
  ***REMOVED***)();
***REMOVED***

function assertStart(value, gen) {
  if (value === GENSYNC_START) return;

  throwError(
    gen,
    makeError(
      `Got unexpected yielded value in gensync generator: ${JSON.stringify(
        value
      )***REMOVED***. Did you perhaps mean to use 'yield*' instead of 'yield'?`,
      GENSYNC_EXPECTED_START
    )
  );
***REMOVED***
function assertSuspend({ value, done ***REMOVED***, gen) {
  if (!done && value === GENSYNC_SUSPEND) return;

  throwError(
    gen,
    makeError(
      done
        ? "Unexpected generator completion. If you get this, it is probably a gensync bug."
        : `Expected GENSYNC_SUSPEND, got ${JSON.stringify(
            value
          )***REMOVED***. If you get this, it is probably a gensync bug.`,
      GENSYNC_EXPECTED_SUSPEND
    )
  );
***REMOVED***

function throwError(gen, err) {
  // Call `.throw` so that users can step in a debugger to easily see which
  // 'yield' passed an unexpected value. If the `.throw` call didn't throw
  // back to the generator, we explicitly do it to stop the error
  // from being swallowed by user code try/catches.
  if (gen.throw) gen.throw(err);
  throw err;
***REMOVED***

function isIterable(value) {
  return (
    !!value &&
    (typeof value === "object" || typeof value === "function") &&
    !value[Symbol.iterator]
  );
***REMOVED***

function setFunctionMetadata(name, arity, fn) {
  if (typeof name === "string") {
    // This should always work on the supported Node versions, but for the
    // sake of users that are compiling to older versions, we check for
    // configurability so we don't throw.
    const nameDesc = Object.getOwnPropertyDescriptor(fn, "name");
    if (!nameDesc || nameDesc.configurable) {
      Object.defineProperty(
        fn,
        "name",
        Object.assign(nameDesc || {***REMOVED***, {
          configurable: true,
          value: name,
        ***REMOVED***)
      );
    ***REMOVED***
  ***REMOVED***

  if (typeof arity === "number") {
    const lengthDesc = Object.getOwnPropertyDescriptor(fn, "length");
    if (!lengthDesc || lengthDesc.configurable) {
      Object.defineProperty(
        fn,
        "length",
        Object.assign(lengthDesc || {***REMOVED***, {
          configurable: true,
          value: arity,
        ***REMOVED***)
      );
    ***REMOVED***
  ***REMOVED***

  return fn;
***REMOVED***
