// postcss-minify-gradients/src/index.js
'use strict';
const valueParser = require('postcss-value-parser');
const { getArguments ***REMOVED*** = require('cssnano-utils');
const isColorStop = require('./isColorStop.js');

const angles = {
  top: '0deg',
  right: '90deg',
  bottom: '180deg',
  left: '270deg',
***REMOVED***;

/**
 * @param {valueParser.Dimension***REMOVED*** a
 * @param {valueParser.Dimension***REMOVED*** b
 * @return {boolean***REMOVED***
 */
function isLessThan(a, b) {
  return (
    a.unit.toLowerCase() === b.unit.toLowerCase() &&
    parseFloat(a.number) >= parseFloat(b.number)
  );
***REMOVED***
/**
 * @param {import('postcss').Declaration***REMOVED*** decl
 * @return {void***REMOVED***
 */
function optimise(decl) {
  const value = decl.value;

  if (!value) {
    return;
  ***REMOVED***

  const normalizedValue = value.toLowerCase();

  if (normalizedValue.includes('var(') || normalizedValue.includes('env(')) {
    return;
  ***REMOVED***

  if (!normalizedValue.includes('gradient')) {
    return;
  ***REMOVED***

  decl.value = valueParser(value)
    .walk((node) => {
      if (node.type !== 'function' || !node.nodes.length) {
        return false;
      ***REMOVED***

      const lowerCasedValue = node.value.toLowerCase();

      if (
        lowerCasedValue === 'linear-gradient' ||
        lowerCasedValue === 'repeating-linear-gradient' ||
        lowerCasedValue === '-webkit-linear-gradient' ||
        lowerCasedValue === '-webkit-repeating-linear-gradient'
      ) {
        let args = getArguments(node);

        if (
          node.nodes[0].value.toLowerCase() === 'to' &&
          args[0].length === 3
        ) {
          node.nodes = node.nodes.slice(2);
          node.nodes[0].value =
            angles[
              /** @type {'top'|'right'|'bottom'|'left'***REMOVED****/ (
                node.nodes[0].value.toLowerCase()
              )
            ];
        ***REMOVED***

        /** @type {valueParser.Dimension | false***REMOVED*** */
        let lastStop;

        args.forEach((arg, index) => {
          if (arg.length !== 3) {
            return;
          ***REMOVED***

          let isFinalStop = index === args.length - 1;
          let thisStop = valueParser.unit(arg[2].value);

          if (lastStop === undefined) {
            lastStop = thisStop;

            if (
              !isFinalStop &&
              lastStop &&
              lastStop.number === '0' &&
              lastStop.unit.toLowerCase() !== 'deg'
            ) {
              arg[1].value = arg[2].value = '';
            ***REMOVED***

            return;
          ***REMOVED***

          if (lastStop && thisStop && isLessThan(lastStop, thisStop)) {
            arg[2].value = '0';
          ***REMOVED***

          lastStop = thisStop;

          if (isFinalStop && arg[2].value === '100%') {
            arg[1].value = arg[2].value = '';
          ***REMOVED***
        ***REMOVED***);

        return false;
      ***REMOVED***

      if (
        lowerCasedValue === 'radial-gradient' ||
        lowerCasedValue === 'repeating-radial-gradient'
      ) {
        let args = getArguments(node);
        /** @type {valueParser.Dimension | false***REMOVED*** */
        let lastStop;

        const hasAt = args[0].find((n) => n.value.toLowerCase() === 'at');

        args.forEach((arg, index) => {
          if (!arg[2] || (!index && hasAt)) {
            return;
          ***REMOVED***

          let thisStop = valueParser.unit(arg[2].value);

          if (!lastStop) {
            lastStop = thisStop;

            return;
          ***REMOVED***

          if (lastStop && thisStop && isLessThan(lastStop, thisStop)) {
            arg[2].value = '0';
          ***REMOVED***

          lastStop = thisStop;
        ***REMOVED***);

        return false;
      ***REMOVED***

      if (
        lowerCasedValue === '-webkit-radial-gradient' ||
        lowerCasedValue === '-webkit-repeating-radial-gradient'
      ) {
        let args = getArguments(node);
        /** @type {valueParser.Dimension | false***REMOVED*** */
        let lastStop;

        args.forEach((arg) => {
          let color;
          let stop;

          if (arg[2] !== undefined) {
            if (arg[0].type === 'function') {
              color = `${arg[0].value***REMOVED***(${valueParser.stringify(arg[0].nodes)***REMOVED***)`;
            ***REMOVED*** else {
              color = arg[0].value;
            ***REMOVED***

            if (arg[2].type === 'function') {
              stop = `${arg[2].value***REMOVED***(${valueParser.stringify(arg[2].nodes)***REMOVED***)`;
            ***REMOVED*** else {
              stop = arg[2].value;
            ***REMOVED***
          ***REMOVED*** else {
            if (arg[0].type === 'function') {
              color = `${arg[0].value***REMOVED***(${valueParser.stringify(arg[0].nodes)***REMOVED***)`;
            ***REMOVED***

            color = arg[0].value;
          ***REMOVED***

          color = color.toLowerCase();

          const colorStop =
            stop !== undefined
              ? isColorStop(color, stop.toLowerCase())
              : isColorStop(color);

          if (!colorStop || !arg[2]) {
            return;
          ***REMOVED***

          let thisStop = valueParser.unit(arg[2].value);

          if (!lastStop) {
            lastStop = thisStop;

            return;
          ***REMOVED***

          if (lastStop && thisStop && isLessThan(lastStop, thisStop)) {
            arg[2].value = '0';
          ***REMOVED***

          lastStop = thisStop;
        ***REMOVED***);

        return false;
      ***REMOVED***
    ***REMOVED***)
    .toString();
***REMOVED***
/**
 * @type {import('postcss').PluginCreator<void>***REMOVED***
 * @return {import('postcss').Plugin***REMOVED***
 */
function pluginCreator() {
  return {
    postcssPlugin: 'postcss-minify-gradients',
    OnceExit(css) {
      css.walkDecls(optimise);
    ***REMOVED***,
  ***REMOVED***;
***REMOVED***

pluginCreator.postcss = true;
module.exports = pluginCreator;
