// @tootallnate/once/dist/index.js
"use strict";
function noop() { ***REMOVED***
function once(emitter, name) {
    const o = once.spread(emitter, name);
    const r = o.then((args) => args[0]);
    r.cancel = o.cancel;
    return r;
***REMOVED***
(function (once) {
    function spread(emitter, name) {
        let c = null;
        const p = new Promise((resolve, reject) => {
            function cancel() {
                emitter.removeListener(name, onEvent);
                emitter.removeListener('error', onError);
                p.cancel = noop;
            ***REMOVED***
            function onEvent(...args) {
                cancel();
                resolve(args);
            ***REMOVED***
            function onError(err) {
                cancel();
                reject(err);
            ***REMOVED***
            c = cancel;
            emitter.on(name, onEvent);
            emitter.on('error', onError);
        ***REMOVED***);
        if (!c) {
            throw new TypeError('Could not get `cancel()` function');
        ***REMOVED***
        p.cancel = c;
        return p;
    ***REMOVED***
    once.spread = spread;
***REMOVED***)(once || (once = {***REMOVED***));
module.exports = once;
