// rollup/dist/shared/watch.js
/*
  @license
	Rollup.js v2.79.2
	Thu, 26 Sep 2024 18:44:14 GMT - commit 48aef33cf2f2a6dfb175afb3bcd6a977c81f1d5c

	https://github.com/rollup/rollup

	Released under the MIT License.
*/
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' ***REMOVED***);

const require$$0 = require('path');
const process = require('process');
const rollup = require('./rollup.js');
const mergeOptions = require('./mergeOptions.js');
const require$$2 = require('os');
const index = require('./index.js');
require('perf_hooks');
require('crypto');
require('fs');
require('events');
require('util');
require('stream');

class FileWatcher {
    constructor(task, chokidarOptions) {
        this.transformWatchers = new Map();
        this.chokidarOptions = chokidarOptions;
        this.task = task;
        this.watcher = this.createWatcher(null);
    ***REMOVED***
    close() {
        this.watcher.close();
        for (const watcher of this.transformWatchers.values()) {
            watcher.close();
        ***REMOVED***
    ***REMOVED***
    unwatch(id) {
        this.watcher.unwatch(id);
        const transformWatcher = this.transformWatchers.get(id);
        if (transformWatcher) {
            this.transformWatchers.delete(id);
            transformWatcher.close();
        ***REMOVED***
    ***REMOVED***
    watch(id, isTransformDependency) {
        var _a;
        if (isTransformDependency) {
            const watcher = (_a = this.transformWatchers.get(id)) !== null && _a !== void 0 ? _a : this.createWatcher(id);
            watcher.add(id);
            this.transformWatchers.set(id, watcher);
        ***REMOVED***
        else {
            this.watcher.add(id);
        ***REMOVED***
    ***REMOVED***
    createWatcher(transformWatcherId) {
        const task = this.task;
        const isLinux = require$$2.platform() === 'linux';
        const isTransformDependency = transformWatcherId !== null;
        const handleChange = (id, event) => {
            const changedId = transformWatcherId || id;
            if (isLinux) {
                // unwatching and watching fixes an issue with chokidar where on certain systems,
                // a file that was unlinked and immediately recreated would create a change event
                // but then no longer any further events
                watcher.unwatch(changedId);
                watcher.add(changedId);
            ***REMOVED***
            task.invalidate(changedId, { event, isTransformDependency ***REMOVED***);
        ***REMOVED***;
        const watcher = index.chokidar
            .watch([], this.chokidarOptions)
            .on('add', id => handleChange(id, 'create'))
            .on('change', id => handleChange(id, 'update'))
            .on('unlink', id => handleChange(id, 'delete'));
        return watcher;
    ***REMOVED***
***REMOVED***

const eventsRewrites = {
    create: {
        create: 'buggy',
        delete: null,
        update: 'create'
    ***REMOVED***,
    delete: {
        create: 'update',
        delete: 'buggy',
        update: 'buggy'
    ***REMOVED***,
    update: {
        create: 'buggy',
        delete: 'delete',
        update: 'update'
    ***REMOVED***
***REMOVED***;
class Watcher {
    constructor(configs, emitter) {
        this.buildDelay = 0;
        this.buildTimeout = null;
        this.invalidatedIds = new Map();
        this.rerun = false;
        this.running = true;
        this.emitter = emitter;
        emitter.close = this.close.bind(this);
        this.tasks = configs.map(config => new Task(this, config));
        this.buildDelay = configs.reduce((buildDelay, { watch ***REMOVED***) => watch && typeof watch.buildDelay === 'number'
            ? Math.max(buildDelay, watch.buildDelay)
            : buildDelay, this.buildDelay);
        process.nextTick(() => this.run());
    ***REMOVED***
    async close() {
        if (this.buildTimeout)
            clearTimeout(this.buildTimeout);
        for (const task of this.tasks) {
            task.close();
        ***REMOVED***
        await this.emitter.emitAndAwait('close');
        this.emitter.removeAllListeners();
    ***REMOVED***
    invalidate(file) {
        if (file) {
            const prevEvent = this.invalidatedIds.get(file.id);
            const event = prevEvent ? eventsRewrites[prevEvent][file.event] : file.event;
            if (event === 'buggy') {
                //TODO: throws or warn? Currently just ignore, uses new event
                this.invalidatedIds.set(file.id, file.event);
            ***REMOVED***
            else if (event === null) {
                this.invalidatedIds.delete(file.id);
            ***REMOVED***
            else {
                this.invalidatedIds.set(file.id, event);
            ***REMOVED***
        ***REMOVED***
        if (this.running) {
            this.rerun = true;
            return;
        ***REMOVED***
        if (this.buildTimeout)
            clearTimeout(this.buildTimeout);
        this.buildTimeout = setTimeout(async () => {
            this.buildTimeout = null;
            try {
                await Promise.all([...this.invalidatedIds].map(([id, event]) => this.emitter.emitAndAwait('change', id, { event ***REMOVED***)));
                this.invalidatedIds.clear();
                this.emitter.emit('restart');
                this.emitter.removeAwaited();
                this.run();
            ***REMOVED***
            catch (error) {
                this.invalidatedIds.clear();
                this.emitter.emit('event', {
                    code: 'ERROR',
                    error,
                    result: null
                ***REMOVED***);
                this.emitter.emit('event', {
                    code: 'END'
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***, this.buildDelay);
    ***REMOVED***
    async run() {
        this.running = true;
        this.emitter.emit('event', {
            code: 'START'
        ***REMOVED***);
        for (const task of this.tasks) {
            await task.run();
        ***REMOVED***
        this.running = false;
        this.emitter.emit('event', {
            code: 'END'
        ***REMOVED***);
        if (this.rerun) {
            this.rerun = false;
            this.invalidate();
        ***REMOVED***
    ***REMOVED***
***REMOVED***
class Task {
    constructor(watcher, config) {
        this.cache = { modules: [] ***REMOVED***;
        this.watchFiles = [];
        this.closed = false;
        this.invalidated = true;
        this.watched = new Set();
        this.watcher = watcher;
        this.skipWrite = Boolean(config.watch && config.watch.skipWrite);
        this.options = mergeOptions.mergeOptions(config);
        this.outputs = this.options.output;
        this.outputFiles = this.outputs.map(output => {
            if (output.file || output.dir)
                return require$$0.resolve(output.file || output.dir);
            return undefined;
        ***REMOVED***);
        const watchOptions = this.options.watch || {***REMOVED***;
        this.filter = rollup.createFilter(watchOptions.include, watchOptions.exclude);
        this.fileWatcher = new FileWatcher(this, {
            ...watchOptions.chokidar,
            disableGlobbing: true,
            ignoreInitial: true
        ***REMOVED***);
    ***REMOVED***
    close() {
        this.closed = true;
        this.fileWatcher.close();
    ***REMOVED***
    invalidate(id, details) {
        this.invalidated = true;
        if (details.isTransformDependency) {
            for (const module of this.cache.modules) {
                if (!module.transformDependencies.includes(id))
                    continue;
                // effective invalidation
                module.originalCode = null;
            ***REMOVED***
        ***REMOVED***
        this.watcher.invalidate({ event: details.event, id ***REMOVED***);
    ***REMOVED***
    async run() {
        if (!this.invalidated)
            return;
        this.invalidated = false;
        const options = {
            ...this.options,
            cache: this.cache
        ***REMOVED***;
        const start = Date.now();
        this.watcher.emitter.emit('event', {
            code: 'BUNDLE_START',
            input: this.options.input,
            output: this.outputFiles
        ***REMOVED***);
        let result = null;
        try {
            result = await rollup.rollupInternal(options, this.watcher.emitter);
            if (this.closed) {
                return;
            ***REMOVED***
            this.updateWatchedFiles(result);
            this.skipWrite || (await Promise.all(this.outputs.map(output => result.write(output))));
            this.watcher.emitter.emit('event', {
                code: 'BUNDLE_END',
                duration: Date.now() - start,
                input: this.options.input,
                output: this.outputFiles,
                result
            ***REMOVED***);
        ***REMOVED***
        catch (error) {
            if (!this.closed) {
                if (Array.isArray(error.watchFiles)) {
                    for (const id of error.watchFiles) {
                        this.watchFile(id);
                    ***REMOVED***
                ***REMOVED***
                if (error.id) {
                    this.cache.modules = this.cache.modules.filter(module => module.id !== error.id);
                ***REMOVED***
            ***REMOVED***
            this.watcher.emitter.emit('event', {
                code: 'ERROR',
                error,
                result
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***
    updateWatchedFiles(result) {
        const previouslyWatched = this.watched;
        this.watched = new Set();
        this.watchFiles = result.watchFiles;
        this.cache = result.cache;
        for (const id of this.watchFiles) {
            this.watchFile(id);
        ***REMOVED***
        for (const module of this.cache.modules) {
            for (const depId of module.transformDependencies) {
                this.watchFile(depId, true);
            ***REMOVED***
        ***REMOVED***
        for (const id of previouslyWatched) {
            if (!this.watched.has(id)) {
                this.fileWatcher.unwatch(id);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    watchFile(id, isTransformDependency = false) {
        if (!this.filter(id))
            return;
        this.watched.add(id);
        if (this.outputFiles.some(file => file === id)) {
            throw new Error('Cannot import the generated bundle');
        ***REMOVED***
        // this is necessary to ensure that any 'renamed' files
        // continue to be watched following an error
        this.fileWatcher.watch(id, isTransformDependency);
    ***REMOVED***
***REMOVED***

exports.Task = Task;
exports.Watcher = Watcher;
