// rollup/dist/shared/watch-cli.js
/*
  @license
	Rollup.js v2.79.2
	Thu, 26 Sep 2024 18:44:14 GMT - commit 48aef33cf2f2a6dfb175afb3bcd6a977c81f1d5c

	https://github.com/rollup/rollup

	Released under the MIT License.
*/
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' ***REMOVED***);

const require$$0$2 = require('fs');
const process$2 = require('process');
const index = require('./index.js');
const cli = require('../bin/rollup');
const rollup = require('./rollup.js');
const require$$0 = require('assert');
const require$$0$1 = require('events');
const loadConfigFile_js = require('./loadConfigFile.js');
const child_process = require('child_process');
require('util');
require('stream');
require('path');
require('os');
require('./mergeOptions.js');
require('perf_hooks');
require('crypto');
require('url');
require('tty');

function timeZone(date = new Date()) {
	const offset = date.getTimezoneOffset();
	const absOffset = Math.abs(offset);
	const hours = Math.floor(absOffset / 60);
	const minutes = absOffset % 60;
	const minutesOut = minutes > 0 ? ':' + ('0' + minutes).slice(-2) : '';
	return (offset < 0 ? '+' : '-') + hours + minutesOut;
***REMOVED***

function dateTime(options = {***REMOVED***) {
	let {
		date = new Date(),
		local = true,
		showTimeZone = false,
		showMilliseconds = false
	***REMOVED*** = options;

	if (local) {
		// Offset the date so it will return the correct value when getting the ISO string.
		date = new Date(date.getTime() - (date.getTimezoneOffset() * 60000));
	***REMOVED***

	let end = '';

	if (showTimeZone) {
		end = ' UTC' + (local ? timeZone(date) : '');
	***REMOVED***

	if (showMilliseconds && date.getUTCMilliseconds() > 0) {
		end = ` ${date.getUTCMilliseconds()***REMOVED***ms${end***REMOVED***`;
	***REMOVED***

	return date
		.toISOString()
		.replace(/T/, ' ')
		.replace(/\..+/, end);
***REMOVED***

var signalExit = {exports: {***REMOVED******REMOVED***;

var signals$1 = {exports: {***REMOVED******REMOVED***;

var hasRequiredSignals;

function requireSignals () {
	if (hasRequiredSignals) return signals$1.exports;
	hasRequiredSignals = 1;
	(function (module) {
		// This is not the set of all possible signals.
		//
		// It IS, however, the set of all signals that trigger
		// an exit on either Linux or BSD systems.  Linux is a
		// superset of the signal names supported on BSD, and
		// the unknown signals just fail to register, so we can
		// catch that easily enough.
		//
		// Don't bother with SIGKILL.  It's uncatchable, which
		// means that we can't fire any callbacks anyway.
		//
		// If a user does happen to register a handler on a non-
		// fatal signal like SIGWINCH or something, and then
		// exit, it'll end up firing `process.emit('exit')`, so
		// the handler will be fired anyway.
		//
		// SIGBUS, SIGFPE, SIGSEGV and SIGILL, when not raised
		// artificially, inherently leave the process in a
		// state from which it is not safe to try and enter JS
		// listeners.
		module.exports = [
		  'SIGABRT',
		  'SIGALRM',
		  'SIGHUP',
		  'SIGINT',
		  'SIGTERM'
		];

		if (process.platform !== 'win32') {
		  module.exports.push(
		    'SIGVTALRM',
		    'SIGXCPU',
		    'SIGXFSZ',
		    'SIGUSR2',
		    'SIGTRAP',
		    'SIGSYS',
		    'SIGQUIT',
		    'SIGIOT'
		    // should detect profiler and enable/disable accordingly.
		    // see #21
		    // 'SIGPROF'
		  );
		***REMOVED***

		if (process.platform === 'linux') {
		  module.exports.push(
		    'SIGIO',
		    'SIGPOLL',
		    'SIGPWR',
		    'SIGSTKFLT',
		    'SIGUNUSED'
		  );
		***REMOVED***
***REMOVED*** (signals$1));
	return signals$1.exports;
***REMOVED***

// Note: since nyc uses this module to output coverage, any lines
// that are in the direct sync flow of nyc's outputCoverage are
// ignored, since we can never get coverage for them.
// grab a reference to node's real process object right away
var process$1 = rollup.commonjsGlobal.process;

const processOk = function (process) {
  return process &&
    typeof process === 'object' &&
    typeof process.removeListener === 'function' &&
    typeof process.emit === 'function' &&
    typeof process.reallyExit === 'function' &&
    typeof process.listeners === 'function' &&
    typeof process.kill === 'function' &&
    typeof process.pid === 'number' &&
    typeof process.on === 'function'
***REMOVED***;

// some kind of non-node environment, just no-op
/* istanbul ignore if */
if (!processOk(process$1)) {
  signalExit.exports = function () {
    return function () {***REMOVED***
  ***REMOVED***;
***REMOVED*** else {
  var assert = require$$0;
  var signals = requireSignals();
  var isWin = /^win/i.test(process$1.platform);

  var EE = require$$0$1;
  /* istanbul ignore if */
  if (typeof EE !== 'function') {
    EE = EE.EventEmitter;
  ***REMOVED***

  var emitter;
  if (process$1.__signal_exit_emitter__) {
    emitter = process$1.__signal_exit_emitter__;
  ***REMOVED*** else {
    emitter = process$1.__signal_exit_emitter__ = new EE();
    emitter.count = 0;
    emitter.emitted = {***REMOVED***;
  ***REMOVED***

  // Because this emitter is a global, we have to check to see if a
  // previous version of this library failed to enable infinite listeners.
  // I know what you're about to say.  But literally everything about
  // signal-exit is a compromise with evil.  Get used to it.
  if (!emitter.infinite) {
    emitter.setMaxListeners(Infinity);
    emitter.infinite = true;
  ***REMOVED***

  signalExit.exports = function (cb, opts) {
    /* istanbul ignore if */
    if (!processOk(rollup.commonjsGlobal.process)) {
      return function () {***REMOVED***
    ***REMOVED***
    assert.equal(typeof cb, 'function', 'a callback must be provided for exit handler');

    if (loaded === false) {
      load();
    ***REMOVED***

    var ev = 'exit';
    if (opts && opts.alwaysLast) {
      ev = 'afterexit';
    ***REMOVED***

    var remove = function () {
      emitter.removeListener(ev, cb);
      if (emitter.listeners('exit').length === 0 &&
          emitter.listeners('afterexit').length === 0) {
        unload();
      ***REMOVED***
    ***REMOVED***;
    emitter.on(ev, cb);

    return remove
  ***REMOVED***;

  var unload = function unload () {
    if (!loaded || !processOk(rollup.commonjsGlobal.process)) {
      return
    ***REMOVED***
    loaded = false;

    signals.forEach(function (sig) {
      try {
        process$1.removeListener(sig, sigListeners[sig]);
      ***REMOVED*** catch (er) {***REMOVED***
    ***REMOVED***);
    process$1.emit = originalProcessEmit;
    process$1.reallyExit = originalProcessReallyExit;
    emitter.count -= 1;
  ***REMOVED***;
  signalExit.exports.unload = unload;

  var emit = function emit (event, code, signal) {
    /* istanbul ignore if */
    if (emitter.emitted[event]) {
      return
    ***REMOVED***
    emitter.emitted[event] = true;
    emitter.emit(event, code, signal);
  ***REMOVED***;

  // { <signal>: <listener fn>, ... ***REMOVED***
  var sigListeners = {***REMOVED***;
  signals.forEach(function (sig) {
    sigListeners[sig] = function listener () {
      /* istanbul ignore if */
      if (!processOk(rollup.commonjsGlobal.process)) {
        return
      ***REMOVED***
      // If there are no other listeners, an exit is coming!
      // Simplest way: remove us and then re-send the signal.
      // We know that this will kill the process, so we can
      // safely emit now.
      var listeners = process$1.listeners(sig);
      if (listeners.length === emitter.count) {
        unload();
        emit('exit', null, sig);
        /* istanbul ignore next */
        emit('afterexit', null, sig);
        /* istanbul ignore next */
        if (isWin && sig === 'SIGHUP') {
          // "SIGHUP" throws an `ENOSYS` error on Windows,
          // so use a supported signal instead
          sig = 'SIGINT';
        ***REMOVED***
        /* istanbul ignore next */
        process$1.kill(process$1.pid, sig);
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***);

  signalExit.exports.signals = function () {
    return signals
  ***REMOVED***;

  var loaded = false;

  var load = function load () {
    if (loaded || !processOk(rollup.commonjsGlobal.process)) {
      return
    ***REMOVED***
    loaded = true;

    // This is the number of onSignalExit's that are in play.
    // It's important so that we can count the correct number of
    // listeners on signals, and don't wait for the other one to
    // handle it instead of us.
    emitter.count += 1;

    signals = signals.filter(function (sig) {
      try {
        process$1.on(sig, sigListeners[sig]);
        return true
      ***REMOVED*** catch (er) {
        return false
      ***REMOVED***
    ***REMOVED***);

    process$1.emit = processEmit;
    process$1.reallyExit = processReallyExit;
  ***REMOVED***;
  signalExit.exports.load = load;

  var originalProcessReallyExit = process$1.reallyExit;
  var processReallyExit = function processReallyExit (code) {
    /* istanbul ignore if */
    if (!processOk(rollup.commonjsGlobal.process)) {
      return
    ***REMOVED***
    process$1.exitCode = code || /* istanbul ignore next */ 0;
    emit('exit', process$1.exitCode, null);
    /* istanbul ignore next */
    emit('afterexit', process$1.exitCode, null);
    /* istanbul ignore next */
    originalProcessReallyExit.call(process$1, process$1.exitCode);
  ***REMOVED***;

  var originalProcessEmit = process$1.emit;
  var processEmit = function processEmit (ev, arg) {
    if (ev === 'exit' && processOk(rollup.commonjsGlobal.process)) {
      /* istanbul ignore else */
      if (arg !== undefined) {
        process$1.exitCode = arg;
      ***REMOVED***
      var ret = originalProcessEmit.apply(this, arguments);
      /* istanbul ignore next */
      emit('exit', process$1.exitCode, null);
      /* istanbul ignore next */
      emit('afterexit', process$1.exitCode, null);
      /* istanbul ignore next */
      return ret
    ***REMOVED*** else {
      return originalProcessEmit.apply(this, arguments)
    ***REMOVED***
  ***REMOVED***;
***REMOVED***

const CLEAR_SCREEN = '\u001Bc';
function getResetScreen(configs, allowClearScreen) {
    let clearScreen = allowClearScreen;
    for (const config of configs) {
        if (config.watch && config.watch.clearScreen === false) {
            clearScreen = false;
        ***REMOVED***
    ***REMOVED***
    if (clearScreen) {
        return (heading) => loadConfigFile_js.stderr(CLEAR_SCREEN + heading);
    ***REMOVED***
    let firstRun = true;
    return (heading) => {
        if (firstRun) {
            loadConfigFile_js.stderr(heading);
            firstRun = false;
        ***REMOVED***
    ***REMOVED***;
***REMOVED***

function extractWatchHooks(command) {
    if (!Array.isArray(command.watch))
        return {***REMOVED***;
    return command.watch
        .filter(value => typeof value === 'object')
        .reduce((acc, keyValueOption) => ({ ...acc, ...keyValueOption ***REMOVED***), {***REMOVED***);
***REMOVED***
function createWatchHooks(command) {
    const watchHooks = extractWatchHooks(command);
    return function (hook) {
        if (watchHooks[hook]) {
            const cmd = watchHooks[hook];
            if (!command.silent) {
                loadConfigFile_js.stderr(loadConfigFile_js.cyan(`watch.${hook***REMOVED*** ${loadConfigFile_js.bold(`$ ${cmd***REMOVED***`)***REMOVED***`));
            ***REMOVED***
            try {
                // !! important - use stderr for all writes from execSync
                const stdio = [process.stdin, process.stderr, process.stderr];
                child_process.execSync(cmd, { stdio: command.silent ? 'ignore' : stdio ***REMOVED***);
            ***REMOVED***
            catch (e) {
                loadConfigFile_js.stderr(e.message);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
***REMOVED***

async function watch(command) {
    process$2.env.ROLLUP_WATCH = 'true';
    const isTTY = process$2.stderr.isTTY;
    const silent = command.silent;
    let watcher;
    let configWatcher;
    let resetScreen;
    const configFile = command.config ? await cli.getConfigPath(command.config) : null;
    const runWatchHook = createWatchHooks(command);
    signalExit.exports(close);
    process$2.on('uncaughtException', close);
    if (!process$2.stdin.isTTY) {
        process$2.stdin.on('end', close);
        process$2.stdin.resume();
    ***REMOVED***
    async function loadConfigFromFileAndTrack(configFile) {
        let configFileData = null;
        let configFileRevision = 0;
        configWatcher = index.chokidar.watch(configFile).on('change', reloadConfigFile);
        await reloadConfigFile();
        async function reloadConfigFile() {
            try {
                const newConfigFileData = await require$$0$2.promises.readFile(configFile, 'utf8');
                if (newConfigFileData === configFileData) {
                    return;
                ***REMOVED***
                configFileRevision++;
                const currentConfigFileRevision = configFileRevision;
                if (configFileData) {
                    loadConfigFile_js.stderr(`\nReloading updated config...`);
                ***REMOVED***
                configFileData = newConfigFileData;
                const { options, warnings ***REMOVED*** = await loadConfigFile_js.loadAndParseConfigFile(configFile, command);
                if (currentConfigFileRevision !== configFileRevision) {
                    return;
                ***REMOVED***
                if (watcher) {
                    await watcher.close();
                ***REMOVED***
                start(options, warnings);
            ***REMOVED***
            catch (err) {
                loadConfigFile_js.handleError(err, true);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    if (configFile) {
        await loadConfigFromFileAndTrack(configFile);
    ***REMOVED***
    else {
        const { options, warnings ***REMOVED*** = await cli.loadConfigFromCommand(command);
        start(options, warnings);
    ***REMOVED***
    function start(configs, warnings) {
        try {
            watcher = rollup.watch(configs);
        ***REMOVED***
        catch (err) {
            return loadConfigFile_js.handleError(err);
        ***REMOVED***
        watcher.on('event', event => {
            switch (event.code) {
                case 'ERROR':
                    warnings.flush();
                    loadConfigFile_js.handleError(event.error, true);
                    runWatchHook('onError');
                    break;
                case 'START':
                    if (!silent) {
                        if (!resetScreen) {
                            resetScreen = getResetScreen(configs, isTTY);
                        ***REMOVED***
                        resetScreen(loadConfigFile_js.underline(`rollup v${rollup.version***REMOVED***`));
                    ***REMOVED***
                    runWatchHook('onStart');
                    break;
                case 'BUNDLE_START':
                    if (!silent) {
                        let input = event.input;
                        if (typeof input !== 'string') {
                            input = Array.isArray(input)
                                ? input.join(', ')
                                : Object.values(input).join(', ');
                        ***REMOVED***
                        loadConfigFile_js.stderr(loadConfigFile_js.cyan(`bundles ${loadConfigFile_js.bold(input)***REMOVED*** → ${loadConfigFile_js.bold(event.output.map(rollup.relativeId).join(', '))***REMOVED***...`));
                    ***REMOVED***
                    runWatchHook('onBundleStart');
                    break;
                case 'BUNDLE_END':
                    warnings.flush();
                    if (!silent)
                        loadConfigFile_js.stderr(loadConfigFile_js.green(`created ${loadConfigFile_js.bold(event.output.map(rollup.relativeId).join(', '))***REMOVED*** in ${loadConfigFile_js.bold(cli.ms(event.duration))***REMOVED***`));
                    runWatchHook('onBundleEnd');
                    if (event.result && event.result.getTimings) {
                        cli.printTimings(event.result.getTimings());
                    ***REMOVED***
                    break;
                case 'END':
                    runWatchHook('onEnd');
                    if (!silent && isTTY) {
                        loadConfigFile_js.stderr(`\n[${dateTime()***REMOVED***] waiting for changes...`);
                    ***REMOVED***
            ***REMOVED***
            if ('result' in event && event.result) {
                event.result.close().catch(error => loadConfigFile_js.handleError(error, true));
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***
    async function close(code) {
        process$2.removeListener('uncaughtException', close);
        // removing a non-existent listener is a no-op
        process$2.stdin.removeListener('end', close);
        if (watcher)
            await watcher.close();
        if (configWatcher)
            configWatcher.close();
        if (code) {
            process$2.exit(code);
        ***REMOVED***
    ***REMOVED***
***REMOVED***

exports.watch = watch;
