// source-list-map/lib/CodeNode.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

const getNumberOfLines = require("./helpers").getNumberOfLines;
const getUnfinishedLine = require("./helpers").getUnfinishedLine;

class CodeNode {
	constructor(generatedCode) {
		this.generatedCode = generatedCode;
	***REMOVED***

	clone() {
		return new CodeNode(this.generatedCode);
	***REMOVED***

	getGeneratedCode() {
		return this.generatedCode;
	***REMOVED***

	getMappings(mappingsContext) {
		const lines = getNumberOfLines(this.generatedCode);
		const mapping = Array(lines+1).join(";");
		if(lines > 0) {
			mappingsContext.unfinishedGeneratedLine = getUnfinishedLine(this.generatedCode);
			if(mappingsContext.unfinishedGeneratedLine > 0) {
				return mapping + "A";
			***REMOVED*** else {
				return mapping;
			***REMOVED***
		***REMOVED*** else {
			const prevUnfinished = mappingsContext.unfinishedGeneratedLine;
			mappingsContext.unfinishedGeneratedLine += getUnfinishedLine(this.generatedCode);
			if(prevUnfinished === 0 && mappingsContext.unfinishedGeneratedLine > 0) {
				return "A";
			***REMOVED*** else {
				return "";
			***REMOVED***
		***REMOVED***
	***REMOVED***

	addGeneratedCode(generatedCode) {
		this.generatedCode += generatedCode;
	***REMOVED***

	mapGeneratedCode(fn) {
		const generatedCode = fn(this.generatedCode);
		return new CodeNode(generatedCode);
	***REMOVED***

	getNormalizedNodes() {
		return [this];
	***REMOVED***

	merge(otherNode) {
		if(otherNode instanceof CodeNode) {
			this.generatedCode += otherNode.generatedCode;
			return this;
		***REMOVED***
		return false;
	***REMOVED***
***REMOVED***

module.exports = CodeNode;
