// workbox-range-requests/src/utils/calculateEffectiveBoundaries.ts
/*
  Copyright 2018 Google LLC

  Use of this source code is governed by an MIT-style
  license that can be found in the LICENSE file or at
  https://opensource.org/licenses/MIT.
*/

import {WorkboxError***REMOVED*** from 'workbox-core/_private/WorkboxError.js';
import {assert***REMOVED*** from 'workbox-core/_private/assert.js';
import '../_version.js';

/**
 * @param {Blob***REMOVED*** blob A source blob.
 * @param {number***REMOVED*** [start] The offset to use as the start of the
 * slice.
 * @param {number***REMOVED*** [end] The offset to use as the end of the slice.
 * @return {Object***REMOVED*** An object with `start` and `end` properties, reflecting
 * the effective boundaries to use given the size of the blob.
 *
 * @private
 */
function calculateEffectiveBoundaries(
  blob: Blob,
  start?: number,
  end?: number,
): {start: number; end: number***REMOVED*** {
  if (process.env.NODE_ENV !== 'production') {
    assert!.isInstance(blob, Blob, {
      moduleName: 'workbox-range-requests',
      funcName: 'calculateEffectiveBoundaries',
      paramName: 'blob',
    ***REMOVED***);
  ***REMOVED***

  const blobSize = blob.size;

  if ((end && end > blobSize) || (start && start < 0)) {
    throw new WorkboxError('range-not-satisfiable', {
      size: blobSize,
      end,
      start,
    ***REMOVED***);
  ***REMOVED***

  let effectiveStart: number;
  let effectiveEnd: number;

  if (start !== undefined && end !== undefined) {
    effectiveStart = start;
    // Range values are inclusive, so add 1 to the value.
    effectiveEnd = end + 1;
  ***REMOVED*** else if (start !== undefined && end === undefined) {
    effectiveStart = start;
    effectiveEnd = blobSize;
  ***REMOVED*** else if (end !== undefined && start === undefined) {
    effectiveStart = blobSize - end;
    effectiveEnd = blobSize;
  ***REMOVED***

  return {
    start: effectiveStart!,
    end: effectiveEnd!,
  ***REMOVED***;
***REMOVED***

export {calculateEffectiveBoundaries***REMOVED***;
