// @babel/plugin-proposal-optional-chaining/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true ***REMOVED***);

var helperPluginUtils = require('@babel/helper-plugin-utils');
var syntaxOptionalChaining = require('@babel/plugin-syntax-optional-chaining');
var core = require('@babel/core');
var helperSkipTransparentExpressionWrappers = require('@babel/helper-skip-transparent-expression-wrappers');

function willPathCastToBoolean(path) {
  const maybeWrapped = findOutermostTransparentParent(path);
  const {
    node,
    parentPath
  ***REMOVED*** = maybeWrapped;
  if (parentPath.isLogicalExpression()) {
    const {
      operator,
      right
    ***REMOVED*** = parentPath.node;
    if (operator === "&&" || operator === "||" || operator === "??" && node === right) {
      return willPathCastToBoolean(parentPath);
    ***REMOVED***
  ***REMOVED***
  if (parentPath.isSequenceExpression()) {
    const {
      expressions
    ***REMOVED*** = parentPath.node;
    if (expressions[expressions.length - 1] === node) {
      return willPathCastToBoolean(parentPath);
    ***REMOVED*** else {
      return true;
    ***REMOVED***
  ***REMOVED***
  return parentPath.isConditional({
    test: node
  ***REMOVED***) || parentPath.isUnaryExpression({
    operator: "!"
  ***REMOVED***) || parentPath.isLoop({
    test: node
  ***REMOVED***);
***REMOVED***
function findOutermostTransparentParent(path) {
  let maybeWrapped = path;
  path.findParent(p => {
    if (!helperSkipTransparentExpressionWrappers.isTransparentExprWrapper(p.node)) return true;
    maybeWrapped = p;
  ***REMOVED***);
  return maybeWrapped;
***REMOVED***

const {
  ast
***REMOVED*** = core.template.expression;
function isSimpleMemberExpression(expression) {
  expression = helperSkipTransparentExpressionWrappers.skipTransparentExprWrapperNodes(expression);
  return core.types.isIdentifier(expression) || core.types.isSuper(expression) || core.types.isMemberExpression(expression) && !expression.computed && isSimpleMemberExpression(expression.object);
***REMOVED***
function needsMemoize(path) {
  let optionalPath = path;
  const {
    scope
  ***REMOVED*** = path;
  while (optionalPath.isOptionalMemberExpression() || optionalPath.isOptionalCallExpression()) {
    const {
      node
    ***REMOVED*** = optionalPath;
    const childPath = helperSkipTransparentExpressionWrappers.skipTransparentExprWrappers(optionalPath.isOptionalMemberExpression() ? optionalPath.get("object") : optionalPath.get("callee"));
    if (node.optional) {
      return !scope.isStatic(childPath.node);
    ***REMOVED***
    optionalPath = childPath;
  ***REMOVED***
***REMOVED***
function transform(path, {
  pureGetters,
  noDocumentAll
***REMOVED***) {
  const {
    scope
  ***REMOVED*** = path;
  const maybeWrapped = findOutermostTransparentParent(path);
  const {
    parentPath
  ***REMOVED*** = maybeWrapped;
  const willReplacementCastToBoolean = willPathCastToBoolean(maybeWrapped);
  let isDeleteOperation = false;
  const parentIsCall = parentPath.isCallExpression({
    callee: maybeWrapped.node
  ***REMOVED***) && path.isOptionalMemberExpression();
  const optionals = [];
  let optionalPath = path;
  if (scope.path.isPattern() && needsMemoize(optionalPath)) {
    path.replaceWith(core.template.ast`(() => ${path.node***REMOVED***)()`);
    return;
  ***REMOVED***
  while (optionalPath.isOptionalMemberExpression() || optionalPath.isOptionalCallExpression()) {
    const {
      node
    ***REMOVED*** = optionalPath;
    if (node.optional) {
      optionals.push(node);
    ***REMOVED***
    if (optionalPath.isOptionalMemberExpression()) {
      optionalPath.node.type = "MemberExpression";
      optionalPath = helperSkipTransparentExpressionWrappers.skipTransparentExprWrappers(optionalPath.get("object"));
    ***REMOVED*** else if (optionalPath.isOptionalCallExpression()) {
      optionalPath.node.type = "CallExpression";
      optionalPath = helperSkipTransparentExpressionWrappers.skipTransparentExprWrappers(optionalPath.get("callee"));
    ***REMOVED***
  ***REMOVED***
  let replacementPath = path;
  if (parentPath.isUnaryExpression({
    operator: "delete"
  ***REMOVED***)) {
    replacementPath = parentPath;
    isDeleteOperation = true;
  ***REMOVED***
  for (let i = optionals.length - 1; i >= 0; i--) {
    const node = optionals[i];
    const isCall = core.types.isCallExpression(node);
    const chainWithTypes = isCall ? node.callee : node.object;
    const chain = helperSkipTransparentExpressionWrappers.skipTransparentExprWrapperNodes(chainWithTypes);
    let ref;
    let check;
    if (isCall && core.types.isIdentifier(chain, {
      name: "eval"
    ***REMOVED***)) {
      check = ref = chain;
      node.callee = core.types.sequenceExpression([core.types.numericLiteral(0), ref]);
    ***REMOVED*** else if (pureGetters && isCall && isSimpleMemberExpression(chain)) {
      check = ref = node.callee;
    ***REMOVED*** else {
      ref = scope.maybeGenerateMemoised(chain);
      if (ref) {
        check = core.types.assignmentExpression("=", core.types.cloneNode(ref), chainWithTypes);
        isCall ? node.callee = ref : node.object = ref;
      ***REMOVED*** else {
        check = ref = chainWithTypes;
      ***REMOVED***
    ***REMOVED***
    if (isCall && core.types.isMemberExpression(chain)) {
      if (pureGetters && isSimpleMemberExpression(chain)) {
        node.callee = chainWithTypes;
      ***REMOVED*** else {
        const {
          object
        ***REMOVED*** = chain;
        let context;
        if (core.types.isSuper(object)) {
          context = core.types.thisExpression();
        ***REMOVED*** else {
          const memoized = scope.maybeGenerateMemoised(object);
          if (memoized) {
            context = memoized;
            chain.object = core.types.assignmentExpression("=", memoized, object);
          ***REMOVED*** else {
            context = object;
          ***REMOVED***
        ***REMOVED***
        node.arguments.unshift(core.types.cloneNode(context));
        node.callee = core.types.memberExpression(node.callee, core.types.identifier("call"));
      ***REMOVED***
    ***REMOVED***
    let replacement = replacementPath.node;
    if (i === 0 && parentIsCall) {
      var _baseRef;
      const object = helperSkipTransparentExpressionWrappers.skipTransparentExprWrapperNodes(replacement.object);
      let baseRef;
      if (!pureGetters || !isSimpleMemberExpression(object)) {
        baseRef = scope.maybeGenerateMemoised(object);
        if (baseRef) {
          replacement.object = core.types.assignmentExpression("=", baseRef, object);
        ***REMOVED***
      ***REMOVED***
      replacement = core.types.callExpression(core.types.memberExpression(replacement, core.types.identifier("bind")), [core.types.cloneNode((_baseRef = baseRef) != null ? _baseRef : object)]);
    ***REMOVED***
    if (willReplacementCastToBoolean) {
      const nonNullishCheck = noDocumentAll ? ast`${core.types.cloneNode(check)***REMOVED*** != null` : ast`
            ${core.types.cloneNode(check)***REMOVED*** !== null && ${core.types.cloneNode(ref)***REMOVED*** !== void 0`;
      replacementPath.replaceWith(core.types.logicalExpression("&&", nonNullishCheck, replacement));
      replacementPath = helperSkipTransparentExpressionWrappers.skipTransparentExprWrappers(replacementPath.get("right"));
    ***REMOVED*** else {
      const nullishCheck = noDocumentAll ? ast`${core.types.cloneNode(check)***REMOVED*** == null` : ast`
            ${core.types.cloneNode(check)***REMOVED*** === null || ${core.types.cloneNode(ref)***REMOVED*** === void 0`;
      const returnValue = isDeleteOperation ? ast`true` : ast`void 0`;
      replacementPath.replaceWith(core.types.conditionalExpression(nullishCheck, returnValue, replacement));
      replacementPath = helperSkipTransparentExpressionWrappers.skipTransparentExprWrappers(replacementPath.get("alternate"));
    ***REMOVED***
  ***REMOVED***
***REMOVED***

var index = helperPluginUtils.declare((api, options) => {
  var _api$assumption, _api$assumption2;
  api.assertVersion(7);
  const {
    loose = false
  ***REMOVED*** = options;
  const noDocumentAll = (_api$assumption = api.assumption("noDocumentAll")) != null ? _api$assumption : loose;
  const pureGetters = (_api$assumption2 = api.assumption("pureGetters")) != null ? _api$assumption2 : loose;
  return {
    name: "proposal-optional-chaining",
    inherits: syntaxOptionalChaining.default,
    visitor: {
      "OptionalCallExpression|OptionalMemberExpression"(path) {
        transform(path, {
          noDocumentAll,
          pureGetters
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;
***REMOVED***);

exports["default"] = index;
exports.transform = transform;
