// ajv/lib/keyword.js
'use strict';

var IDENTIFIER = /^[a-z_$][a-z0-9_$-]*$/i;
var customRuleCode = require('./dotjs/custom');
var definitionSchema = require('./definition_schema');

module.exports = {
  add: addKeyword,
  get: getKeyword,
  remove: removeKeyword,
  validate: validateKeyword
***REMOVED***;


/**
 * Define custom keyword
 * @this  Ajv
 * @param {String***REMOVED*** keyword custom keyword, should be unique (including different from all standard, custom and macro keywords).
 * @param {Object***REMOVED*** definition keyword definition object with properties `type` (type(s) which the keyword applies to), `validate` or `compile`.
 * @return {Ajv***REMOVED*** this for method chaining
 */
function addKeyword(keyword, definition) {
  /* jshint validthis: true */
  /* eslint no-shadow: 0 */
  var RULES = this.RULES;
  if (RULES.keywords[keyword])
    throw new Error('Keyword ' + keyword + ' is already defined');

  if (!IDENTIFIER.test(keyword))
    throw new Error('Keyword ' + keyword + ' is not a valid identifier');

  if (definition) {
    this.validateKeyword(definition, true);

    var dataType = definition.type;
    if (Array.isArray(dataType)) {
      for (var i=0; i<dataType.length; i++)
        _addRule(keyword, dataType[i], definition);
    ***REMOVED*** else {
      _addRule(keyword, dataType, definition);
    ***REMOVED***

    var metaSchema = definition.metaSchema;
    if (metaSchema) {
      if (definition.$data && this._opts.$data) {
        metaSchema = {
          anyOf: [
            metaSchema,
            { '$ref': 'https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#' ***REMOVED***
          ]
        ***REMOVED***;
      ***REMOVED***
      definition.validateSchema = this.compile(metaSchema, true);
    ***REMOVED***
  ***REMOVED***

  RULES.keywords[keyword] = RULES.all[keyword] = true;


  function _addRule(keyword, dataType, definition) {
    var ruleGroup;
    for (var i=0; i<RULES.length; i++) {
      var rg = RULES[i];
      if (rg.type == dataType) {
        ruleGroup = rg;
        break;
      ***REMOVED***
    ***REMOVED***

    if (!ruleGroup) {
      ruleGroup = { type: dataType, rules: [] ***REMOVED***;
      RULES.push(ruleGroup);
    ***REMOVED***

    var rule = {
      keyword: keyword,
      definition: definition,
      custom: true,
      code: customRuleCode,
      implements: definition.implements
    ***REMOVED***;
    ruleGroup.rules.push(rule);
    RULES.custom[keyword] = rule;
  ***REMOVED***

  return this;
***REMOVED***


/**
 * Get keyword
 * @this  Ajv
 * @param {String***REMOVED*** keyword pre-defined or custom keyword.
 * @return {Object|Boolean***REMOVED*** custom keyword definition, `true` if it is a predefined keyword, `false` otherwise.
 */
function getKeyword(keyword) {
  /* jshint validthis: true */
  var rule = this.RULES.custom[keyword];
  return rule ? rule.definition : this.RULES.keywords[keyword] || false;
***REMOVED***


/**
 * Remove keyword
 * @this  Ajv
 * @param {String***REMOVED*** keyword pre-defined or custom keyword.
 * @return {Ajv***REMOVED*** this for method chaining
 */
function removeKeyword(keyword) {
  /* jshint validthis: true */
  var RULES = this.RULES;
  delete RULES.keywords[keyword];
  delete RULES.all[keyword];
  delete RULES.custom[keyword];
  for (var i=0; i<RULES.length; i++) {
    var rules = RULES[i].rules;
    for (var j=0; j<rules.length; j++) {
      if (rules[j].keyword == keyword) {
        rules.splice(j, 1);
        break;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  return this;
***REMOVED***


/**
 * Validate keyword definition
 * @this  Ajv
 * @param {Object***REMOVED*** definition keyword definition object.
 * @param {Boolean***REMOVED*** throwError true to throw exception if definition is invalid
 * @return {boolean***REMOVED*** validation result
 */
function validateKeyword(definition, throwError) {
  validateKeyword.errors = null;
  var v = this._validateKeyword = this._validateKeyword
                                  || this.compile(definitionSchema, true);

  if (v(definition)) return true;
  validateKeyword.errors = v.errors;
  if (throwError)
    throw new Error('custom keyword definition is invalid: '  + this.errorsText(v.errors));
  else
    return false;
***REMOVED***
