// exit/Gruntfile.js
'use strict';

module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    nodeunit: {
      files: ['test/**/*_test.js'],
    ***REMOVED***,
    jshint: {
      options: {
        jshintrc: '.jshintrc'
      ***REMOVED***,
      gruntfile: {
        src: 'Gruntfile.js'
      ***REMOVED***,
      lib: {
        src: ['lib/**/*.js']
      ***REMOVED***,
      test: {
        src: ['test/**/*.js']
      ***REMOVED***,
    ***REMOVED***,
    watch: {
      gruntfile: {
        files: '<%= jshint.gruntfile.src %>',
        tasks: ['jshint:gruntfile']
      ***REMOVED***,
      lib: {
        files: '<%= jshint.lib.src %>',
        tasks: ['jshint:lib', 'nodeunit']
      ***REMOVED***,
      test: {
        files: '<%= jshint.test.src %>',
        tasks: ['jshint:test', 'nodeunit']
      ***REMOVED***,
    ***REMOVED***,
  ***REMOVED***);

  // These plugins provide necessary tasks.
  grunt.loadNpmTasks('grunt-contrib-nodeunit');
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-watch');

  // Default task.
  grunt.registerTask('default', ['jshint', 'nodeunit']);

***REMOVED***;
