// eslint/lib/source-code/token-store/cursors.js
/**
 * @fileoverview Define 2 token factories; forward and backward.
 * @author Toru Nagashima
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const BackwardTokenCommentCursor = require("./backward-token-comment-cursor");
const BackwardTokenCursor = require("./backward-token-cursor");
const FilterCursor = require("./filter-cursor");
const ForwardTokenCommentCursor = require("./forward-token-comment-cursor");
const ForwardTokenCursor = require("./forward-token-cursor");
const LimitCursor = require("./limit-cursor");
const SkipCursor = require("./skip-cursor");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * The cursor factory.
 * @private
 */
class CursorFactory {

    /**
     * Initializes this cursor.
     * @param {Function***REMOVED*** TokenCursor The class of the cursor which iterates tokens only.
     * @param {Function***REMOVED*** TokenCommentCursor The class of the cursor which iterates the mix of tokens and comments.
     */
    constructor(TokenCursor, TokenCommentCursor) {
        this.TokenCursor = TokenCursor;
        this.TokenCommentCursor = TokenCommentCursor;
    ***REMOVED***

    /**
     * Creates a base cursor instance that can be decorated by createCursor.
     * @param {Token[]***REMOVED*** tokens The array of tokens.
     * @param {Comment[]***REMOVED*** comments The array of comments.
     * @param {Object***REMOVED*** indexMap The map from locations to indices in `tokens`.
     * @param {number***REMOVED*** startLoc The start location of the iteration range.
     * @param {number***REMOVED*** endLoc The end location of the iteration range.
     * @param {boolean***REMOVED*** includeComments The flag to iterate comments as well.
     * @returns {Cursor***REMOVED*** The created base cursor.
     */
    createBaseCursor(tokens, comments, indexMap, startLoc, endLoc, includeComments) {
        const Cursor = includeComments ? this.TokenCommentCursor : this.TokenCursor;

        return new Cursor(tokens, comments, indexMap, startLoc, endLoc);
    ***REMOVED***

    /**
     * Creates a cursor that iterates tokens with normalized options.
     * @param {Token[]***REMOVED*** tokens The array of tokens.
     * @param {Comment[]***REMOVED*** comments The array of comments.
     * @param {Object***REMOVED*** indexMap The map from locations to indices in `tokens`.
     * @param {number***REMOVED*** startLoc The start location of the iteration range.
     * @param {number***REMOVED*** endLoc The end location of the iteration range.
     * @param {boolean***REMOVED*** includeComments The flag to iterate comments as well.
     * @param {Function|null***REMOVED*** filter The predicate function to choose tokens.
     * @param {number***REMOVED*** skip The count of tokens the cursor skips.
     * @param {number***REMOVED*** count The maximum count of tokens the cursor iterates. Zero is no iteration for backward compatibility.
     * @returns {Cursor***REMOVED*** The created cursor.
     */
    createCursor(tokens, comments, indexMap, startLoc, endLoc, includeComments, filter, skip, count) {
        let cursor = this.createBaseCursor(tokens, comments, indexMap, startLoc, endLoc, includeComments);

        if (filter) {
            cursor = new FilterCursor(cursor, filter);
        ***REMOVED***
        if (skip >= 1) {
            cursor = new SkipCursor(cursor, skip);
        ***REMOVED***
        if (count >= 0) {
            cursor = new LimitCursor(cursor, count);
        ***REMOVED***

        return cursor;
    ***REMOVED***
***REMOVED***

//------------------------------------------------------------------------------
// Exports
//------------------------------------------------------------------------------

exports.forward = new CursorFactory(ForwardTokenCursor, ForwardTokenCommentCursor);
exports.backward = new CursorFactory(BackwardTokenCursor, BackwardTokenCommentCursor);
