// cssstyle/lib/CSSStyleDeclaration.js
/*********************************************************************
 * This is a fork from the CSS Style Declaration part of
 * https://github.com/NV/CSSOM
 ********************************************************************/
'use strict';
var CSSOM = require('cssom');
var allProperties = require('./allProperties');
var allExtraProperties = require('./allExtraProperties');
var implementedProperties = require('./implementedProperties');
var { dashedToCamelCase ***REMOVED*** = require('./parsers');
var getBasicPropertyDescriptor = require('./utils/getBasicPropertyDescriptor');

/**
 * @constructor
 * @see http://www.w3.org/TR/DOM-Level-2-Style/css.html#CSS-CSSStyleDeclaration
 */
var CSSStyleDeclaration = function CSSStyleDeclaration(onChangeCallback) {
  this._values = {***REMOVED***;
  this._importants = {***REMOVED***;
  this._length = 0;
  this._onChange =
    onChangeCallback ||
    function() {
      return;
    ***REMOVED***;
***REMOVED***;
CSSStyleDeclaration.prototype = {
  constructor: CSSStyleDeclaration,

  /**
   *
   * @param {string***REMOVED*** name
   * @see http://www.w3.org/TR/DOM-Level-2-Style/css.html#CSS-CSSStyleDeclaration-getPropertyValue
   * @return {string***REMOVED*** the value of the property if it has been explicitly set for this declaration block.
   * Returns the empty string if the property has not been set.
   */
  getPropertyValue: function(name) {
    if (!this._values.hasOwnProperty(name)) {
      return '';
    ***REMOVED***
    return this._values[name].toString();
  ***REMOVED***,

  /**
   *
   * @param {string***REMOVED*** name
   * @param {string***REMOVED*** value
   * @param {string***REMOVED*** [priority=null] "important" or null
   * @see http://www.w3.org/TR/DOM-Level-2-Style/css.html#CSS-CSSStyleDeclaration-setProperty
   */
  setProperty: function(name, value, priority) {
    if (value === undefined) {
      return;
    ***REMOVED***
    if (value === null || value === '') {
      this.removeProperty(name);
      return;
    ***REMOVED***
    var isCustomProperty = name.indexOf('--') === 0;
    if (isCustomProperty) {
      this._setProperty(name, value, priority);
      return;
    ***REMOVED***
    var lowercaseName = name.toLowerCase();
    if (!allProperties.has(lowercaseName) && !allExtraProperties.has(lowercaseName)) {
      return;
    ***REMOVED***

    this[lowercaseName] = value;
    this._importants[lowercaseName] = priority;
  ***REMOVED***,
  _setProperty: function(name, value, priority) {
    if (value === undefined) {
      return;
    ***REMOVED***
    if (value === null || value === '') {
      this.removeProperty(name);
      return;
    ***REMOVED***
    if (this._values[name]) {
      // Property already exist. Overwrite it.
      var index = Array.prototype.indexOf.call(this, name);
      if (index < 0) {
        this[this._length] = name;
        this._length++;
      ***REMOVED***
    ***REMOVED*** else {
      // New property.
      this[this._length] = name;
      this._length++;
    ***REMOVED***
    this._values[name] = value;
    this._importants[name] = priority;
    this._onChange(this.cssText);
  ***REMOVED***,

  /**
   *
   * @param {string***REMOVED*** name
   * @see http://www.w3.org/TR/DOM-Level-2-Style/css.html#CSS-CSSStyleDeclaration-removeProperty
   * @return {string***REMOVED*** the value of the property if it has been explicitly set for this declaration block.
   * Returns the empty string if the property has not been set or the property name does not correspond to a known CSS property.
   */
  removeProperty: function(name) {
    if (!this._values.hasOwnProperty(name)) {
      return '';
    ***REMOVED***

    var prevValue = this._values[name];
    delete this._values[name];
    delete this._importants[name];

    var index = Array.prototype.indexOf.call(this, name);
    if (index < 0) {
      return prevValue;
    ***REMOVED***

    // That's what WebKit and Opera do
    Array.prototype.splice.call(this, index, 1);

    // That's what Firefox does
    //this[index] = ""

    this._onChange(this.cssText);
    return prevValue;
  ***REMOVED***,

  /**
   *
   * @param {String***REMOVED*** name
   */
  getPropertyPriority: function(name) {
    return this._importants[name] || '';
  ***REMOVED***,

  getPropertyCSSValue: function() {
    //FIXME
    return;
  ***REMOVED***,

  /**
   *   element.style.overflow = "auto"
   *   element.style.getPropertyShorthand("overflow-x")
   *   -> "overflow"
   */
  getPropertyShorthand: function() {
    //FIXME
    return;
  ***REMOVED***,

  isPropertyImplicit: function() {
    //FIXME
    return;
  ***REMOVED***,

  /**
   *   http://www.w3.org/TR/DOM-Level-2-Style/css.html#CSS-CSSStyleDeclaration-item
   */
  item: function(index) {
    index = parseInt(index, 10);
    if (index < 0 || index >= this._length) {
      return '';
    ***REMOVED***
    return this[index];
  ***REMOVED***,
***REMOVED***;

Object.defineProperties(CSSStyleDeclaration.prototype, {
  cssText: {
    get: function() {
      var properties = [];
      var i;
      var name;
      var value;
      var priority;
      for (i = 0; i < this._length; i++) {
        name = this[i];
        value = this.getPropertyValue(name);
        priority = this.getPropertyPriority(name);
        if (priority !== '') {
          priority = ' !' + priority;
        ***REMOVED***
        properties.push([name, ': ', value, priority, ';'].join(''));
      ***REMOVED***
      return properties.join(' ');
    ***REMOVED***,
    set: function(value) {
      var i;
      this._values = {***REMOVED***;
      Array.prototype.splice.call(this, 0, this._length);
      this._importants = {***REMOVED***;
      var dummyRule;
      try {
        dummyRule = CSSOM.parse('#bogus{' + value + '***REMOVED***').cssRules[0].style;
      ***REMOVED*** catch (err) {
        // malformed css, just return
        return;
      ***REMOVED***
      var rule_length = dummyRule.length;
      var name;
      for (i = 0; i < rule_length; ++i) {
        name = dummyRule[i];
        this.setProperty(
          dummyRule[i],
          dummyRule.getPropertyValue(name),
          dummyRule.getPropertyPriority(name)
        );
      ***REMOVED***
      this._onChange(this.cssText);
    ***REMOVED***,
    enumerable: true,
    configurable: true,
  ***REMOVED***,
  parentRule: {
    get: function() {
      return null;
    ***REMOVED***,
    enumerable: true,
    configurable: true,
  ***REMOVED***,
  length: {
    get: function() {
      return this._length;
    ***REMOVED***,
    /**
     * This deletes indices if the new length is less then the current
     * length. If the new length is more, it does nothing, the new indices
     * will be undefined until set.
     **/
    set: function(value) {
      var i;
      for (i = value; i < this._length; i++) {
        delete this[i];
      ***REMOVED***
      this._length = value;
    ***REMOVED***,
    enumerable: true,
    configurable: true,
  ***REMOVED***,
***REMOVED***);

require('./properties')(CSSStyleDeclaration.prototype);

allProperties.forEach(function(property) {
  if (!implementedProperties.has(property)) {
    var declaration = getBasicPropertyDescriptor(property);
    Object.defineProperty(CSSStyleDeclaration.prototype, property, declaration);
    Object.defineProperty(CSSStyleDeclaration.prototype, dashedToCamelCase(property), declaration);
  ***REMOVED***
***REMOVED***);

allExtraProperties.forEach(function(property) {
  if (!implementedProperties.has(property)) {
    var declaration = getBasicPropertyDescriptor(property);
    Object.defineProperty(CSSStyleDeclaration.prototype, property, declaration);
    Object.defineProperty(CSSStyleDeclaration.prototype, dashedToCamelCase(property), declaration);
  ***REMOVED***
***REMOVED***);

exports.CSSStyleDeclaration = CSSStyleDeclaration;
