// schema-utils/dist/ValidationError.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

const {
  stringHints,
  numberHints
***REMOVED*** = require('./util/hints');
/** @typedef {import("json-schema").JSONSchema6***REMOVED*** JSONSchema6 */

/** @typedef {import("json-schema").JSONSchema7***REMOVED*** JSONSchema7 */

/** @typedef {import("./validate").Schema***REMOVED*** Schema */

/** @typedef {import("./validate").ValidationErrorConfiguration***REMOVED*** ValidationErrorConfiguration */

/** @typedef {import("./validate").PostFormatter***REMOVED*** PostFormatter */

/** @typedef {import("./validate").SchemaUtilErrorObject***REMOVED*** SchemaUtilErrorObject */

/** @enum {number***REMOVED*** */


const SPECIFICITY = {
  type: 1,
  not: 1,
  oneOf: 1,
  anyOf: 1,
  if: 1,
  enum: 1,
  const: 1,
  instanceof: 1,
  required: 2,
  pattern: 2,
  patternRequired: 2,
  format: 2,
  formatMinimum: 2,
  formatMaximum: 2,
  minimum: 2,
  exclusiveMinimum: 2,
  maximum: 2,
  exclusiveMaximum: 2,
  multipleOf: 2,
  uniqueItems: 2,
  contains: 2,
  minLength: 2,
  maxLength: 2,
  minItems: 2,
  maxItems: 2,
  minProperties: 2,
  maxProperties: 2,
  dependencies: 2,
  propertyNames: 2,
  additionalItems: 2,
  additionalProperties: 2,
  absolutePath: 2
***REMOVED***;
/**
 *
 * @param {Array<SchemaUtilErrorObject>***REMOVED*** array
 * @param {(item: SchemaUtilErrorObject) => number***REMOVED*** fn
 * @returns {Array<SchemaUtilErrorObject>***REMOVED***
 */

function filterMax(array, fn) {
  const evaluatedMax = array.reduce((max, item) => Math.max(max, fn(item)), 0);
  return array.filter(item => fn(item) === evaluatedMax);
***REMOVED***
/**
 *
 * @param {Array<SchemaUtilErrorObject>***REMOVED*** children
 * @returns {Array<SchemaUtilErrorObject>***REMOVED***
 */


function filterChildren(children) {
  let newChildren = children;
  newChildren = filterMax(newChildren,
  /**
   *
   * @param {SchemaUtilErrorObject***REMOVED*** error
   * @returns {number***REMOVED***
   */
  error => error.dataPath ? error.dataPath.length : 0);
  newChildren = filterMax(newChildren,
  /**
   * @param {SchemaUtilErrorObject***REMOVED*** error
   * @returns {number***REMOVED***
   */
  error => SPECIFICITY[
  /** @type {keyof typeof SPECIFICITY***REMOVED*** */
  error.keyword] || 2);
  return newChildren;
***REMOVED***
/**
 * Find all children errors
 * @param {Array<SchemaUtilErrorObject>***REMOVED*** children
 * @param {Array<string>***REMOVED*** schemaPaths
 * @return {number***REMOVED*** returns index of first child
 */


function findAllChildren(children, schemaPaths) {
  let i = children.length - 1;

  const predicate =
  /**
   * @param {string***REMOVED*** schemaPath
   * @returns {boolean***REMOVED***
   */
  schemaPath => children[i].schemaPath.indexOf(schemaPath) !== 0;

  while (i > -1 && !schemaPaths.every(predicate)) {
    if (children[i].keyword === 'anyOf' || children[i].keyword === 'oneOf') {
      const refs = extractRefs(children[i]);
      const childrenStart = findAllChildren(children.slice(0, i), refs.concat(children[i].schemaPath));
      i = childrenStart - 1;
    ***REMOVED*** else {
      i -= 1;
    ***REMOVED***
  ***REMOVED***

  return i + 1;
***REMOVED***
/**
 * Extracts all refs from schema
 * @param {SchemaUtilErrorObject***REMOVED*** error
 * @return {Array<string>***REMOVED***
 */


function extractRefs(error) {
  const {
    schema
  ***REMOVED*** = error;

  if (!Array.isArray(schema)) {
    return [];
  ***REMOVED***

  return schema.map(({
    $ref
  ***REMOVED***) => $ref).filter(s => s);
***REMOVED***
/**
 * Groups children by their first level parent (assuming that error is root)
 * @param {Array<SchemaUtilErrorObject>***REMOVED*** children
 * @return {Array<SchemaUtilErrorObject>***REMOVED***
 */


function groupChildrenByFirstChild(children) {
  const result = [];
  let i = children.length - 1;

  while (i > 0) {
    const child = children[i];

    if (child.keyword === 'anyOf' || child.keyword === 'oneOf') {
      const refs = extractRefs(child);
      const childrenStart = findAllChildren(children.slice(0, i), refs.concat(child.schemaPath));

      if (childrenStart !== i) {
        result.push(Object.assign({***REMOVED***, child, {
          children: children.slice(childrenStart, i)
        ***REMOVED***));
        i = childrenStart;
      ***REMOVED*** else {
        result.push(child);
      ***REMOVED***
    ***REMOVED*** else {
      result.push(child);
    ***REMOVED***

    i -= 1;
  ***REMOVED***

  if (i === 0) {
    result.push(children[i]);
  ***REMOVED***

  return result.reverse();
***REMOVED***
/**
 * @param {string***REMOVED*** str
 * @param {string***REMOVED*** prefix
 * @returns {string***REMOVED***
 */


function indent(str, prefix) {
  return str.replace(/\n(?!$)/g, `\n${prefix***REMOVED***`);
***REMOVED***
/**
 * @param {Schema***REMOVED*** schema
 * @returns {schema is (Schema & {not: Schema***REMOVED***)***REMOVED***
 */


function hasNotInSchema(schema) {
  return !!schema.not;
***REMOVED***
/**
 * @param {Schema***REMOVED*** schema
 * @return {Schema***REMOVED***
 */


function findFirstTypedSchema(schema) {
  if (hasNotInSchema(schema)) {
    return findFirstTypedSchema(schema.not);
  ***REMOVED***

  return schema;
***REMOVED***
/**
 * @param {Schema***REMOVED*** schema
 * @return {boolean***REMOVED***
 */


function canApplyNot(schema) {
  const typedSchema = findFirstTypedSchema(schema);
  return likeNumber(typedSchema) || likeInteger(typedSchema) || likeString(typedSchema) || likeNull(typedSchema) || likeBoolean(typedSchema);
***REMOVED***
/**
 * @param {any***REMOVED*** maybeObj
 * @returns {boolean***REMOVED***
 */


function isObject(maybeObj) {
  return typeof maybeObj === 'object' && maybeObj !== null;
***REMOVED***
/**
 * @param {Schema***REMOVED*** schema
 * @returns {boolean***REMOVED***
 */


function likeNumber(schema) {
  return schema.type === 'number' || typeof schema.minimum !== 'undefined' || typeof schema.exclusiveMinimum !== 'undefined' || typeof schema.maximum !== 'undefined' || typeof schema.exclusiveMaximum !== 'undefined' || typeof schema.multipleOf !== 'undefined';
***REMOVED***
/**
 * @param {Schema***REMOVED*** schema
 * @returns {boolean***REMOVED***
 */


function likeInteger(schema) {
  return schema.type === 'integer' || typeof schema.minimum !== 'undefined' || typeof schema.exclusiveMinimum !== 'undefined' || typeof schema.maximum !== 'undefined' || typeof schema.exclusiveMaximum !== 'undefined' || typeof schema.multipleOf !== 'undefined';
***REMOVED***
/**
 * @param {Schema***REMOVED*** schema
 * @returns {boolean***REMOVED***
 */


function likeString(schema) {
  return schema.type === 'string' || typeof schema.minLength !== 'undefined' || typeof schema.maxLength !== 'undefined' || typeof schema.pattern !== 'undefined' || typeof schema.format !== 'undefined' || typeof schema.formatMinimum !== 'undefined' || typeof schema.formatMaximum !== 'undefined';
***REMOVED***
/**
 * @param {Schema***REMOVED*** schema
 * @returns {boolean***REMOVED***
 */


function likeBoolean(schema) {
  return schema.type === 'boolean';
***REMOVED***
/**
 * @param {Schema***REMOVED*** schema
 * @returns {boolean***REMOVED***
 */


function likeArray(schema) {
  return schema.type === 'array' || typeof schema.minItems === 'number' || typeof schema.maxItems === 'number' || typeof schema.uniqueItems !== 'undefined' || typeof schema.items !== 'undefined' || typeof schema.additionalItems !== 'undefined' || typeof schema.contains !== 'undefined';
***REMOVED***
/**
 * @param {Schema & {patternRequired?: Array<string>***REMOVED******REMOVED*** schema
 * @returns {boolean***REMOVED***
 */


function likeObject(schema) {
  return schema.type === 'object' || typeof schema.minProperties !== 'undefined' || typeof schema.maxProperties !== 'undefined' || typeof schema.required !== 'undefined' || typeof schema.properties !== 'undefined' || typeof schema.patternProperties !== 'undefined' || typeof schema.additionalProperties !== 'undefined' || typeof schema.dependencies !== 'undefined' || typeof schema.propertyNames !== 'undefined' || typeof schema.patternRequired !== 'undefined';
***REMOVED***
/**
 * @param {Schema***REMOVED*** schema
 * @returns {boolean***REMOVED***
 */


function likeNull(schema) {
  return schema.type === 'null';
***REMOVED***
/**
 * @param {string***REMOVED*** type
 * @returns {string***REMOVED***
 */


function getArticle(type) {
  if (/^[aeiou]/i.test(type)) {
    return 'an';
  ***REMOVED***

  return 'a';
***REMOVED***
/**
 * @param {Schema=***REMOVED*** schema
 * @returns {string***REMOVED***
 */


function getSchemaNonTypes(schema) {
  if (!schema) {
    return '';
  ***REMOVED***

  if (!schema.type) {
    if (likeNumber(schema) || likeInteger(schema)) {
      return ' | should be any non-number';
    ***REMOVED***

    if (likeString(schema)) {
      return ' | should be any non-string';
    ***REMOVED***

    if (likeArray(schema)) {
      return ' | should be any non-array';
    ***REMOVED***

    if (likeObject(schema)) {
      return ' | should be any non-object';
    ***REMOVED***
  ***REMOVED***

  return '';
***REMOVED***
/**
 * @param {Array<string>***REMOVED*** hints
 * @returns {string***REMOVED***
 */


function formatHints(hints) {
  return hints.length > 0 ? `(${hints.join(', ')***REMOVED***)` : '';
***REMOVED***
/**
 * @param {Schema***REMOVED*** schema
 * @param {boolean***REMOVED*** logic
 * @returns {string[]***REMOVED***
 */


function getHints(schema, logic) {
  if (likeNumber(schema) || likeInteger(schema)) {
    return numberHints(schema, logic);
  ***REMOVED*** else if (likeString(schema)) {
    return stringHints(schema, logic);
  ***REMOVED***

  return [];
***REMOVED***

class ValidationError extends Error {
  /**
   * @param {Array<SchemaUtilErrorObject>***REMOVED*** errors
   * @param {Schema***REMOVED*** schema
   * @param {ValidationErrorConfiguration***REMOVED*** configuration
   */
  constructor(errors, schema, configuration = {***REMOVED***) {
    super();
    /** @type {string***REMOVED*** */

    this.name = 'ValidationError';
    /** @type {Array<SchemaUtilErrorObject>***REMOVED*** */

    this.errors = errors;
    /** @type {Schema***REMOVED*** */

    this.schema = schema;
    let headerNameFromSchema;
    let baseDataPathFromSchema;

    if (schema.title && (!configuration.name || !configuration.baseDataPath)) {
      const splittedTitleFromSchema = schema.title.match(/^(.+) (.+)$/);

      if (splittedTitleFromSchema) {
        if (!configuration.name) {
          [, headerNameFromSchema] = splittedTitleFromSchema;
        ***REMOVED***

        if (!configuration.baseDataPath) {
          [,, baseDataPathFromSchema] = splittedTitleFromSchema;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
    /** @type {string***REMOVED*** */


    this.headerName = configuration.name || headerNameFromSchema || 'Object';
    /** @type {string***REMOVED*** */

    this.baseDataPath = configuration.baseDataPath || baseDataPathFromSchema || 'configuration';
    /** @type {PostFormatter | null***REMOVED*** */

    this.postFormatter = configuration.postFormatter || null;
    const header = `Invalid ${this.baseDataPath***REMOVED*** object. ${this.headerName***REMOVED*** has been initialized using ${getArticle(this.baseDataPath)***REMOVED*** ${this.baseDataPath***REMOVED*** object that does not match the API schema.\n`;
    /** @type {string***REMOVED*** */

    this.message = `${header***REMOVED***${this.formatValidationErrors(errors)***REMOVED***`;
    Error.captureStackTrace(this, this.constructor);
  ***REMOVED***
  /**
   * @param {string***REMOVED*** path
   * @returns {Schema***REMOVED***
   */


  getSchemaPart(path) {
    const newPath = path.split('/');
    let schemaPart = this.schema;

    for (let i = 1; i < newPath.length; i++) {
      const inner = schemaPart[
      /** @type {keyof Schema***REMOVED*** */
      newPath[i]];

      if (!inner) {
        break;
      ***REMOVED***

      schemaPart = inner;
    ***REMOVED***

    return schemaPart;
  ***REMOVED***
  /**
   * @param {Schema***REMOVED*** schema
   * @param {boolean***REMOVED*** logic
   * @param {Array<Object>***REMOVED*** prevSchemas
   * @returns {string***REMOVED***
   */


  formatSchema(schema, logic = true, prevSchemas = []) {
    let newLogic = logic;

    const formatInnerSchema =
    /**
     *
     * @param {Object***REMOVED*** innerSchema
     * @param {boolean=***REMOVED*** addSelf
     * @returns {string***REMOVED***
     */
    (innerSchema, addSelf) => {
      if (!addSelf) {
        return this.formatSchema(innerSchema, newLogic, prevSchemas);
      ***REMOVED***

      if (prevSchemas.includes(innerSchema)) {
        return '(recursive)';
      ***REMOVED***

      return this.formatSchema(innerSchema, newLogic, prevSchemas.concat(schema));
    ***REMOVED***;

    if (hasNotInSchema(schema) && !likeObject(schema)) {
      if (canApplyNot(schema.not)) {
        newLogic = !logic;
        return formatInnerSchema(schema.not);
      ***REMOVED***

      const needApplyLogicHere = !schema.not.not;
      const prefix = logic ? '' : 'non ';
      newLogic = !logic;
      return needApplyLogicHere ? prefix + formatInnerSchema(schema.not) : formatInnerSchema(schema.not);
    ***REMOVED***

    if (
    /** @type {Schema & {instanceof: string | Array<string>***REMOVED******REMOVED*** */
    schema.instanceof) {
      const {
        instanceof: value
      ***REMOVED*** =
      /** @type {Schema & {instanceof: string | Array<string>***REMOVED******REMOVED*** */
      schema;
      const values = !Array.isArray(value) ? [value] : value;
      return values.map(
      /**
       * @param {string***REMOVED*** item
       * @returns {string***REMOVED***
       */
      item => item === 'Function' ? 'function' : item).join(' | ');
    ***REMOVED***

    if (schema.enum) {
      return (
        /** @type {Array<any>***REMOVED*** */
        schema.enum.map(item => JSON.stringify(item)).join(' | ')
      );
    ***REMOVED***

    if (typeof schema.const !== 'undefined') {
      return JSON.stringify(schema.const);
    ***REMOVED***

    if (schema.oneOf) {
      return (
        /** @type {Array<Schema>***REMOVED*** */
        schema.oneOf.map(item => formatInnerSchema(item, true)).join(' | ')
      );
    ***REMOVED***

    if (schema.anyOf) {
      return (
        /** @type {Array<Schema>***REMOVED*** */
        schema.anyOf.map(item => formatInnerSchema(item, true)).join(' | ')
      );
    ***REMOVED***

    if (schema.allOf) {
      return (
        /** @type {Array<Schema>***REMOVED*** */
        schema.allOf.map(item => formatInnerSchema(item, true)).join(' & ')
      );
    ***REMOVED***

    if (
    /** @type {JSONSchema7***REMOVED*** */
    schema.if) {
      const {
        if: ifValue,
        then: thenValue,
        else: elseValue
      ***REMOVED*** =
      /** @type {JSONSchema7***REMOVED*** */
      schema;
      return `${ifValue ? `if ${formatInnerSchema(ifValue)***REMOVED***` : ''***REMOVED***${thenValue ? ` then ${formatInnerSchema(thenValue)***REMOVED***` : ''***REMOVED***${elseValue ? ` else ${formatInnerSchema(elseValue)***REMOVED***` : ''***REMOVED***`;
    ***REMOVED***

    if (schema.$ref) {
      return formatInnerSchema(this.getSchemaPart(schema.$ref), true);
    ***REMOVED***

    if (likeNumber(schema) || likeInteger(schema)) {
      const [type, ...hints] = getHints(schema, logic);
      const str = `${type***REMOVED***${hints.length > 0 ? ` ${formatHints(hints)***REMOVED***` : ''***REMOVED***`;
      return logic ? str : hints.length > 0 ? `non-${type***REMOVED*** | ${str***REMOVED***` : `non-${type***REMOVED***`;
    ***REMOVED***

    if (likeString(schema)) {
      const [type, ...hints] = getHints(schema, logic);
      const str = `${type***REMOVED***${hints.length > 0 ? ` ${formatHints(hints)***REMOVED***` : ''***REMOVED***`;
      return logic ? str : str === 'string' ? 'non-string' : `non-string | ${str***REMOVED***`;
    ***REMOVED***

    if (likeBoolean(schema)) {
      return `${logic ? '' : 'non-'***REMOVED***boolean`;
    ***REMOVED***

    if (likeArray(schema)) {
      // not logic already applied in formatValidationError
      newLogic = true;
      const hints = [];

      if (typeof schema.minItems === 'number') {
        hints.push(`should not have fewer than ${schema.minItems***REMOVED*** item${schema.minItems > 1 ? 's' : ''***REMOVED***`);
      ***REMOVED***

      if (typeof schema.maxItems === 'number') {
        hints.push(`should not have more than ${schema.maxItems***REMOVED*** item${schema.maxItems > 1 ? 's' : ''***REMOVED***`);
      ***REMOVED***

      if (schema.uniqueItems) {
        hints.push('should not have duplicate items');
      ***REMOVED***

      const hasAdditionalItems = typeof schema.additionalItems === 'undefined' || Boolean(schema.additionalItems);
      let items = '';

      if (schema.items) {
        if (Array.isArray(schema.items) && schema.items.length > 0) {
          items = `${
          /** @type {Array<Schema>***REMOVED*** */
          schema.items.map(item => formatInnerSchema(item)).join(', ')***REMOVED***`;

          if (hasAdditionalItems) {
            if (schema.additionalItems && isObject(schema.additionalItems) && Object.keys(schema.additionalItems).length > 0) {
              hints.push(`additional items should be ${formatInnerSchema(schema.additionalItems)***REMOVED***`);
            ***REMOVED***
          ***REMOVED***
        ***REMOVED*** else if (schema.items && Object.keys(schema.items).length > 0) {
          // "additionalItems" is ignored
          items = `${formatInnerSchema(schema.items)***REMOVED***`;
        ***REMOVED*** else {
          // Fallback for empty `items` value
          items = 'any';
        ***REMOVED***
      ***REMOVED*** else {
        // "additionalItems" is ignored
        items = 'any';
      ***REMOVED***

      if (schema.contains && Object.keys(schema.contains).length > 0) {
        hints.push(`should contains at least one ${this.formatSchema(schema.contains)***REMOVED*** item`);
      ***REMOVED***

      return `[${items***REMOVED***${hasAdditionalItems ? ', ...' : ''***REMOVED***]${hints.length > 0 ? ` (${hints.join(', ')***REMOVED***)` : ''***REMOVED***`;
    ***REMOVED***

    if (likeObject(schema)) {
      // not logic already applied in formatValidationError
      newLogic = true;
      const hints = [];

      if (typeof schema.minProperties === 'number') {
        hints.push(`should not have fewer than ${schema.minProperties***REMOVED*** ${schema.minProperties > 1 ? 'properties' : 'property'***REMOVED***`);
      ***REMOVED***

      if (typeof schema.maxProperties === 'number') {
        hints.push(`should not have more than ${schema.maxProperties***REMOVED*** ${schema.minProperties && schema.minProperties > 1 ? 'properties' : 'property'***REMOVED***`);
      ***REMOVED***

      if (schema.patternProperties && Object.keys(schema.patternProperties).length > 0) {
        const patternProperties = Object.keys(schema.patternProperties);
        hints.push(`additional property names should match pattern${patternProperties.length > 1 ? 's' : ''***REMOVED*** ${patternProperties.map(pattern => JSON.stringify(pattern)).join(' | ')***REMOVED***`);
      ***REMOVED***

      const properties = schema.properties ? Object.keys(schema.properties) : [];
      const required = schema.required ? schema.required : [];
      const allProperties = [...new Set(
      /** @type {Array<string>***REMOVED*** */
      [].concat(required).concat(properties))];
      const objectStructure = allProperties.map(property => {
        const isRequired = required.includes(property); // Some properties need quotes, maybe we should add check
        // Maybe we should output type of property (`foo: string`), but it is looks very unreadable

        return `${property***REMOVED***${isRequired ? '' : '?'***REMOVED***`;
      ***REMOVED***).concat(typeof schema.additionalProperties === 'undefined' || Boolean(schema.additionalProperties) ? schema.additionalProperties && isObject(schema.additionalProperties) ? [`<key>: ${formatInnerSchema(schema.additionalProperties)***REMOVED***`] : ['…'] : []).join(', ');
      const {
        dependencies,
        propertyNames,
        patternRequired
      ***REMOVED*** =
      /** @type {Schema & {patternRequired?: Array<string>;***REMOVED******REMOVED*** */
      schema;

      if (dependencies) {
        Object.keys(dependencies).forEach(dependencyName => {
          const dependency = dependencies[dependencyName];

          if (Array.isArray(dependency)) {
            hints.push(`should have ${dependency.length > 1 ? 'properties' : 'property'***REMOVED*** ${dependency.map(dep => `'${dep***REMOVED***'`).join(', ')***REMOVED*** when property '${dependencyName***REMOVED***' is present`);
          ***REMOVED*** else {
            hints.push(`should be valid according to the schema ${formatInnerSchema(dependency)***REMOVED*** when property '${dependencyName***REMOVED***' is present`);
          ***REMOVED***
        ***REMOVED***);
      ***REMOVED***

      if (propertyNames && Object.keys(propertyNames).length > 0) {
        hints.push(`each property name should match format ${JSON.stringify(schema.propertyNames.format)***REMOVED***`);
      ***REMOVED***

      if (patternRequired && patternRequired.length > 0) {
        hints.push(`should have property matching pattern ${patternRequired.map(
        /**
         * @param {string***REMOVED*** item
         * @returns {string***REMOVED***
         */
        item => JSON.stringify(item))***REMOVED***`);
      ***REMOVED***

      return `object {${objectStructure ? ` ${objectStructure***REMOVED*** ` : ''***REMOVED******REMOVED***${hints.length > 0 ? ` (${hints.join(', ')***REMOVED***)` : ''***REMOVED***`;
    ***REMOVED***

    if (likeNull(schema)) {
      return `${logic ? '' : 'non-'***REMOVED***null`;
    ***REMOVED***

    if (Array.isArray(schema.type)) {
      // not logic already applied in formatValidationError
      return `${schema.type.join(' | ')***REMOVED***`;
    ***REMOVED*** // Fallback for unknown keywords
    // not logic already applied in formatValidationError

    /* istanbul ignore next */


    return JSON.stringify(schema, null, 2);
  ***REMOVED***
  /**
   * @param {Schema=***REMOVED*** schemaPart
   * @param {(boolean | Array<string>)=***REMOVED*** additionalPath
   * @param {boolean=***REMOVED*** needDot
   * @param {boolean=***REMOVED*** logic
   * @returns {string***REMOVED***
   */


  getSchemaPartText(schemaPart, additionalPath, needDot = false, logic = true) {
    if (!schemaPart) {
      return '';
    ***REMOVED***

    if (Array.isArray(additionalPath)) {
      for (let i = 0; i < additionalPath.length; i++) {
        /** @type {Schema | undefined***REMOVED*** */
        const inner = schemaPart[
        /** @type {keyof Schema***REMOVED*** */
        additionalPath[i]];

        if (inner) {
          // eslint-disable-next-line no-param-reassign
          schemaPart = inner;
        ***REMOVED*** else {
          break;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    while (schemaPart.$ref) {
      // eslint-disable-next-line no-param-reassign
      schemaPart = this.getSchemaPart(schemaPart.$ref);
    ***REMOVED***

    let schemaText = `${this.formatSchema(schemaPart, logic)***REMOVED***${needDot ? '.' : ''***REMOVED***`;

    if (schemaPart.description) {
      schemaText += `\n-> ${schemaPart.description***REMOVED***`;
    ***REMOVED***

    return schemaText;
  ***REMOVED***
  /**
   * @param {Schema=***REMOVED*** schemaPart
   * @returns {string***REMOVED***
   */


  getSchemaPartDescription(schemaPart) {
    if (!schemaPart) {
      return '';
    ***REMOVED***

    while (schemaPart.$ref) {
      // eslint-disable-next-line no-param-reassign
      schemaPart = this.getSchemaPart(schemaPart.$ref);
    ***REMOVED***

    if (schemaPart.description) {
      return `\n-> ${schemaPart.description***REMOVED***`;
    ***REMOVED***

    return '';
  ***REMOVED***
  /**
   * @param {SchemaUtilErrorObject***REMOVED*** error
   * @returns {string***REMOVED***
   */


  formatValidationError(error) {
    const {
      keyword,
      dataPath: errorDataPath
    ***REMOVED*** = error;
    const dataPath = `${this.baseDataPath***REMOVED***${errorDataPath***REMOVED***`;

    switch (keyword) {
      case 'type':
        {
          const {
            parentSchema,
            params
          ***REMOVED*** = error; // eslint-disable-next-line default-case

          switch (
          /** @type {import("ajv").TypeParams***REMOVED*** */
          params.type) {
            case 'number':
              return `${dataPath***REMOVED*** should be a ${this.getSchemaPartText(parentSchema, false, true)***REMOVED***`;

            case 'integer':
              return `${dataPath***REMOVED*** should be a ${this.getSchemaPartText(parentSchema, false, true)***REMOVED***`;

            case 'string':
              return `${dataPath***REMOVED*** should be a ${this.getSchemaPartText(parentSchema, false, true)***REMOVED***`;

            case 'boolean':
              return `${dataPath***REMOVED*** should be a ${this.getSchemaPartText(parentSchema, false, true)***REMOVED***`;

            case 'array':
              return `${dataPath***REMOVED*** should be an array:\n${this.getSchemaPartText(parentSchema)***REMOVED***`;

            case 'object':
              return `${dataPath***REMOVED*** should be an object:\n${this.getSchemaPartText(parentSchema)***REMOVED***`;

            case 'null':
              return `${dataPath***REMOVED*** should be a ${this.getSchemaPartText(parentSchema, false, true)***REMOVED***`;

            default:
              return `${dataPath***REMOVED*** should be:\n${this.getSchemaPartText(parentSchema)***REMOVED***`;
          ***REMOVED***
        ***REMOVED***

      case 'instanceof':
        {
          const {
            parentSchema
          ***REMOVED*** = error;
          return `${dataPath***REMOVED*** should be an instance of ${this.getSchemaPartText(parentSchema, false, true)***REMOVED***`;
        ***REMOVED***

      case 'pattern':
        {
          const {
            params,
            parentSchema
          ***REMOVED*** = error;
          const {
            pattern
          ***REMOVED*** =
          /** @type {import("ajv").PatternParams***REMOVED*** */
          params;
          return `${dataPath***REMOVED*** should match pattern ${JSON.stringify(pattern)***REMOVED***${getSchemaNonTypes(parentSchema)***REMOVED***.${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'format':
        {
          const {
            params,
            parentSchema
          ***REMOVED*** = error;
          const {
            format
          ***REMOVED*** =
          /** @type {import("ajv").FormatParams***REMOVED*** */
          params;
          return `${dataPath***REMOVED*** should match format ${JSON.stringify(format)***REMOVED***${getSchemaNonTypes(parentSchema)***REMOVED***.${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'formatMinimum':
      case 'formatMaximum':
        {
          const {
            params,
            parentSchema
          ***REMOVED*** = error;
          const {
            comparison,
            limit
          ***REMOVED*** =
          /** @type {import("ajv").ComparisonParams***REMOVED*** */
          params;
          return `${dataPath***REMOVED*** should be ${comparison***REMOVED*** ${JSON.stringify(limit)***REMOVED***${getSchemaNonTypes(parentSchema)***REMOVED***.${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'minimum':
      case 'maximum':
      case 'exclusiveMinimum':
      case 'exclusiveMaximum':
        {
          const {
            parentSchema,
            params
          ***REMOVED*** = error;
          const {
            comparison,
            limit
          ***REMOVED*** =
          /** @type {import("ajv").ComparisonParams***REMOVED*** */
          params;
          const [, ...hints] = getHints(
          /** @type {Schema***REMOVED*** */
          parentSchema, true);

          if (hints.length === 0) {
            hints.push(`should be ${comparison***REMOVED*** ${limit***REMOVED***`);
          ***REMOVED***

          return `${dataPath***REMOVED*** ${hints.join(' ')***REMOVED***${getSchemaNonTypes(parentSchema)***REMOVED***.${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'multipleOf':
        {
          const {
            params,
            parentSchema
          ***REMOVED*** = error;
          const {
            multipleOf
          ***REMOVED*** =
          /** @type {import("ajv").MultipleOfParams***REMOVED*** */
          params;
          return `${dataPath***REMOVED*** should be multiple of ${multipleOf***REMOVED***${getSchemaNonTypes(parentSchema)***REMOVED***.${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'patternRequired':
        {
          const {
            params,
            parentSchema
          ***REMOVED*** = error;
          const {
            missingPattern
          ***REMOVED*** =
          /** @type {import("ajv").PatternRequiredParams***REMOVED*** */
          params;
          return `${dataPath***REMOVED*** should have property matching pattern ${JSON.stringify(missingPattern)***REMOVED***${getSchemaNonTypes(parentSchema)***REMOVED***.${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'minLength':
        {
          const {
            params,
            parentSchema
          ***REMOVED*** = error;
          const {
            limit
          ***REMOVED*** =
          /** @type {import("ajv").LimitParams***REMOVED*** */
          params;

          if (limit === 1) {
            return `${dataPath***REMOVED*** should be an non-empty string${getSchemaNonTypes(parentSchema)***REMOVED***.${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
          ***REMOVED***

          const length = limit - 1;
          return `${dataPath***REMOVED*** should be longer than ${length***REMOVED*** character${length > 1 ? 's' : ''***REMOVED***${getSchemaNonTypes(parentSchema)***REMOVED***.${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'minItems':
        {
          const {
            params,
            parentSchema
          ***REMOVED*** = error;
          const {
            limit
          ***REMOVED*** =
          /** @type {import("ajv").LimitParams***REMOVED*** */
          params;

          if (limit === 1) {
            return `${dataPath***REMOVED*** should be an non-empty array${getSchemaNonTypes(parentSchema)***REMOVED***.${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
          ***REMOVED***

          return `${dataPath***REMOVED*** should not have fewer than ${limit***REMOVED*** items${getSchemaNonTypes(parentSchema)***REMOVED***.${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'minProperties':
        {
          const {
            params,
            parentSchema
          ***REMOVED*** = error;
          const {
            limit
          ***REMOVED*** =
          /** @type {import("ajv").LimitParams***REMOVED*** */
          params;

          if (limit === 1) {
            return `${dataPath***REMOVED*** should be an non-empty object${getSchemaNonTypes(parentSchema)***REMOVED***.${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
          ***REMOVED***

          return `${dataPath***REMOVED*** should not have fewer than ${limit***REMOVED*** properties${getSchemaNonTypes(parentSchema)***REMOVED***.${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'maxLength':
        {
          const {
            params,
            parentSchema
          ***REMOVED*** = error;
          const {
            limit
          ***REMOVED*** =
          /** @type {import("ajv").LimitParams***REMOVED*** */
          params;
          const max = limit + 1;
          return `${dataPath***REMOVED*** should be shorter than ${max***REMOVED*** character${max > 1 ? 's' : ''***REMOVED***${getSchemaNonTypes(parentSchema)***REMOVED***.${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'maxItems':
        {
          const {
            params,
            parentSchema
          ***REMOVED*** = error;
          const {
            limit
          ***REMOVED*** =
          /** @type {import("ajv").LimitParams***REMOVED*** */
          params;
          return `${dataPath***REMOVED*** should not have more than ${limit***REMOVED*** items${getSchemaNonTypes(parentSchema)***REMOVED***.${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'maxProperties':
        {
          const {
            params,
            parentSchema
          ***REMOVED*** = error;
          const {
            limit
          ***REMOVED*** =
          /** @type {import("ajv").LimitParams***REMOVED*** */
          params;
          return `${dataPath***REMOVED*** should not have more than ${limit***REMOVED*** properties${getSchemaNonTypes(parentSchema)***REMOVED***.${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'uniqueItems':
        {
          const {
            params,
            parentSchema
          ***REMOVED*** = error;
          const {
            i
          ***REMOVED*** =
          /** @type {import("ajv").UniqueItemsParams***REMOVED*** */
          params;
          return `${dataPath***REMOVED*** should not contain the item '${error.data[i]***REMOVED***' twice${getSchemaNonTypes(parentSchema)***REMOVED***.${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'additionalItems':
        {
          const {
            params,
            parentSchema
          ***REMOVED*** = error;
          const {
            limit
          ***REMOVED*** =
          /** @type {import("ajv").LimitParams***REMOVED*** */
          params;
          return `${dataPath***REMOVED*** should not have more than ${limit***REMOVED*** items${getSchemaNonTypes(parentSchema)***REMOVED***. These items are valid:\n${this.getSchemaPartText(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'contains':
        {
          const {
            parentSchema
          ***REMOVED*** = error;
          return `${dataPath***REMOVED*** should contains at least one ${this.getSchemaPartText(parentSchema, ['contains'])***REMOVED*** item${getSchemaNonTypes(parentSchema)***REMOVED***.`;
        ***REMOVED***

      case 'required':
        {
          const {
            parentSchema,
            params
          ***REMOVED*** = error;
          const missingProperty =
          /** @type {import("ajv").DependenciesParams***REMOVED*** */
          params.missingProperty.replace(/^\./, '');
          const hasProperty = parentSchema && Boolean(
          /** @type {Schema***REMOVED*** */
          parentSchema.properties &&
          /** @type {Schema***REMOVED*** */
          parentSchema.properties[missingProperty]);
          return `${dataPath***REMOVED*** misses the property '${missingProperty***REMOVED***'${getSchemaNonTypes(parentSchema)***REMOVED***.${hasProperty ? ` Should be:\n${this.getSchemaPartText(parentSchema, ['properties', missingProperty])***REMOVED***` : this.getSchemaPartDescription(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'additionalProperties':
        {
          const {
            params,
            parentSchema
          ***REMOVED*** = error;
          const {
            additionalProperty
          ***REMOVED*** =
          /** @type {import("ajv").AdditionalPropertiesParams***REMOVED*** */
          params;
          return `${dataPath***REMOVED*** has an unknown property '${additionalProperty***REMOVED***'${getSchemaNonTypes(parentSchema)***REMOVED***. These properties are valid:\n${this.getSchemaPartText(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'dependencies':
        {
          const {
            params,
            parentSchema
          ***REMOVED*** = error;
          const {
            property,
            deps
          ***REMOVED*** =
          /** @type {import("ajv").DependenciesParams***REMOVED*** */
          params;
          const dependencies = deps.split(',').map(
          /**
           * @param {string***REMOVED*** dep
           * @returns {string***REMOVED***
           */
          dep => `'${dep.trim()***REMOVED***'`).join(', ');
          return `${dataPath***REMOVED*** should have properties ${dependencies***REMOVED*** when property '${property***REMOVED***' is present${getSchemaNonTypes(parentSchema)***REMOVED***.${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'propertyNames':
        {
          const {
            params,
            parentSchema,
            schema
          ***REMOVED*** = error;
          const {
            propertyName
          ***REMOVED*** =
          /** @type {import("ajv").PropertyNamesParams***REMOVED*** */
          params;
          return `${dataPath***REMOVED*** property name '${propertyName***REMOVED***' is invalid${getSchemaNonTypes(parentSchema)***REMOVED***. Property names should be match format ${JSON.stringify(schema.format)***REMOVED***.${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'enum':
        {
          const {
            parentSchema
          ***REMOVED*** = error;

          if (parentSchema &&
          /** @type {Schema***REMOVED*** */
          parentSchema.enum &&
          /** @type {Schema***REMOVED*** */
          parentSchema.enum.length === 1) {
            return `${dataPath***REMOVED*** should be ${this.getSchemaPartText(parentSchema, false, true)***REMOVED***`;
          ***REMOVED***

          return `${dataPath***REMOVED*** should be one of these:\n${this.getSchemaPartText(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'const':
        {
          const {
            parentSchema
          ***REMOVED*** = error;
          return `${dataPath***REMOVED*** should be equal to constant ${this.getSchemaPartText(parentSchema, false, true)***REMOVED***`;
        ***REMOVED***

      case 'not':
        {
          const postfix = likeObject(
          /** @type {Schema***REMOVED*** */
          error.parentSchema) ? `\n${this.getSchemaPartText(error.parentSchema)***REMOVED***` : '';
          const schemaOutput = this.getSchemaPartText(error.schema, false, false, false);

          if (canApplyNot(error.schema)) {
            return `${dataPath***REMOVED*** should be any ${schemaOutput***REMOVED***${postfix***REMOVED***.`;
          ***REMOVED***

          const {
            schema,
            parentSchema
          ***REMOVED*** = error;
          return `${dataPath***REMOVED*** should not be ${this.getSchemaPartText(schema, false, true)***REMOVED***${parentSchema && likeObject(parentSchema) ? `\n${this.getSchemaPartText(parentSchema)***REMOVED***` : ''***REMOVED***`;
        ***REMOVED***

      case 'oneOf':
      case 'anyOf':
        {
          const {
            parentSchema,
            children
          ***REMOVED*** = error;

          if (children && children.length > 0) {
            if (error.schema.length === 1) {
              const lastChild = children[children.length - 1];
              const remainingChildren = children.slice(0, children.length - 1);
              return this.formatValidationError(Object.assign({***REMOVED***, lastChild, {
                children: remainingChildren,
                parentSchema: Object.assign({***REMOVED***, parentSchema, lastChild.parentSchema)
              ***REMOVED***));
            ***REMOVED***

            let filteredChildren = filterChildren(children);

            if (filteredChildren.length === 1) {
              return this.formatValidationError(filteredChildren[0]);
            ***REMOVED***

            filteredChildren = groupChildrenByFirstChild(filteredChildren);
            return `${dataPath***REMOVED*** should be one of these:\n${this.getSchemaPartText(parentSchema)***REMOVED***\nDetails:\n${filteredChildren.map(
            /**
             * @param {SchemaUtilErrorObject***REMOVED*** nestedError
             * @returns {string***REMOVED***
             */
            nestedError => ` * ${indent(this.formatValidationError(nestedError), '   ')***REMOVED***`).join('\n')***REMOVED***`;
          ***REMOVED***

          return `${dataPath***REMOVED*** should be one of these:\n${this.getSchemaPartText(parentSchema)***REMOVED***`;
        ***REMOVED***

      case 'if':
        {
          const {
            params,
            parentSchema
          ***REMOVED*** = error;
          const {
            failingKeyword
          ***REMOVED*** =
          /** @type {import("ajv").IfParams***REMOVED*** */
          params;
          return `${dataPath***REMOVED*** should match "${failingKeyword***REMOVED***" schema:\n${this.getSchemaPartText(parentSchema, [failingKeyword])***REMOVED***`;
        ***REMOVED***

      case 'absolutePath':
        {
          const {
            message,
            parentSchema
          ***REMOVED*** = error;
          return `${dataPath***REMOVED***: ${message***REMOVED***${this.getSchemaPartDescription(parentSchema)***REMOVED***`;
        ***REMOVED***

      /* istanbul ignore next */

      default:
        {
          const {
            message,
            parentSchema
          ***REMOVED*** = error;
          const ErrorInJSON = JSON.stringify(error, null, 2); // For `custom`, `false schema`, `$ref` keywords
          // Fallback for unknown keywords

          return `${dataPath***REMOVED*** ${message***REMOVED*** (${ErrorInJSON***REMOVED***).\n${this.getSchemaPartText(parentSchema, false)***REMOVED***`;
        ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  /**
   * @param {Array<SchemaUtilErrorObject>***REMOVED*** errors
   * @returns {string***REMOVED***
   */


  formatValidationErrors(errors) {
    return errors.map(error => {
      let formattedError = this.formatValidationError(error);

      if (this.postFormatter) {
        formattedError = this.postFormatter(formattedError, error);
      ***REMOVED***

      return ` - ${indent(formattedError, '   ')***REMOVED***`;
    ***REMOVED***).join('\n');
  ***REMOVED***

***REMOVED***

var _default = ValidationError;
exports.default = _default;
