// @typescript-eslint/scope-manager/dist/referencer/ExportVisitor.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
***REMOVED***;
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
***REMOVED***;
var _ExportVisitor_referencer, _ExportVisitor_exportNode;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.ExportVisitor = void 0;
const types_1 = require("@typescript-eslint/types");
const Visitor_1 = require("./Visitor");
class ExportVisitor extends Visitor_1.Visitor {
    constructor(node, referencer) {
        super(referencer);
        _ExportVisitor_referencer.set(this, void 0);
        _ExportVisitor_exportNode.set(this, void 0);
        __classPrivateFieldSet(this, _ExportVisitor_exportNode, node, "f");
        __classPrivateFieldSet(this, _ExportVisitor_referencer, referencer, "f");
    ***REMOVED***
    static visit(referencer, node) {
        const exportReferencer = new ExportVisitor(node, referencer);
        exportReferencer.visit(node);
    ***REMOVED***
    Identifier(node) {
        if (__classPrivateFieldGet(this, _ExportVisitor_exportNode, "f").exportKind === 'type') {
            // export type { T ***REMOVED***;
            // type exports can only reference types
            __classPrivateFieldGet(this, _ExportVisitor_referencer, "f").currentScope().referenceType(node);
        ***REMOVED***
        else {
            __classPrivateFieldGet(this, _ExportVisitor_referencer, "f").currentScope().referenceDualValueType(node);
        ***REMOVED***
    ***REMOVED***
    ExportDefaultDeclaration(node) {
        if (node.declaration.type === types_1.AST_NODE_TYPES.Identifier) {
            // export default A;
            // this could be a type or a variable
            this.visit(node.declaration);
        ***REMOVED***
        else {
            // export const a = 1;
            // export something();
            // etc
            // these not included in the scope of this visitor as they are all guaranteed to be values or declare variables
        ***REMOVED***
    ***REMOVED***
    ExportNamedDeclaration(node) {
        if (node.source) {
            // export ... from 'foo';
            // these are external identifiers so there shouldn't be references or defs
            return;
        ***REMOVED***
        if (!node.declaration) {
            // export { x ***REMOVED***;
            this.visitChildren(node);
        ***REMOVED***
        else {
            // export const x = 1;
            // this is not included in the scope of this visitor as it creates a variable
        ***REMOVED***
    ***REMOVED***
    ExportSpecifier(node) {
        if (node.exportKind === 'type') {
            // export { type T ***REMOVED***;
            // type exports can only reference types
            //
            // we can't let this fall through to the Identifier selector because the exportKind is on this node
            // and we don't have access to the `.parent` during scope analysis
            __classPrivateFieldGet(this, _ExportVisitor_referencer, "f").currentScope().referenceType(node.local);
        ***REMOVED***
        else {
            this.visit(node.local);
        ***REMOVED***
    ***REMOVED***
***REMOVED***
exports.ExportVisitor = ExportVisitor;
_ExportVisitor_referencer = new WeakMap(), _ExportVisitor_exportNode = new WeakMap();
