// istanbul-reports/lib/html/insertion-text.js
'use strict';
/*
 Copyright 2012-2015, Yahoo Inc.
 Copyrights licensed under the New BSD License. See the accompanying LICENSE file for terms.
 */
function InsertionText(text, consumeBlanks) {
    this.text = text;
    this.origLength = text.length;
    this.offsets = [];
    this.consumeBlanks = consumeBlanks;
    this.startPos = this.findFirstNonBlank();
    this.endPos = this.findLastNonBlank();
***REMOVED***

const WHITE_RE = /[ \f\n\r\t\v\u00A0\u2028\u2029]/;

InsertionText.prototype = {
    findFirstNonBlank() {
        let pos = -1;
        const text = this.text;
        const len = text.length;
        let i;
        for (i = 0; i < len; i += 1) {
            if (!text.charAt(i).match(WHITE_RE)) {
                pos = i;
                break;
            ***REMOVED***
        ***REMOVED***
        return pos;
    ***REMOVED***,
    findLastNonBlank() {
        const text = this.text;
        const len = text.length;
        let pos = text.length + 1;
        let i;
        for (i = len - 1; i >= 0; i -= 1) {
            if (!text.charAt(i).match(WHITE_RE)) {
                pos = i;
                break;
            ***REMOVED***
        ***REMOVED***
        return pos;
    ***REMOVED***,
    originalLength() {
        return this.origLength;
    ***REMOVED***,

    insertAt(col, str, insertBefore, consumeBlanks) {
        consumeBlanks =
            typeof consumeBlanks === 'undefined'
                ? this.consumeBlanks
                : consumeBlanks;
        col = col > this.originalLength() ? this.originalLength() : col;
        col = col < 0 ? 0 : col;

        if (consumeBlanks) {
            if (col <= this.startPos) {
                col = 0;
            ***REMOVED***
            if (col > this.endPos) {
                col = this.origLength;
            ***REMOVED***
        ***REMOVED***

        const len = str.length;
        const offset = this.findOffset(col, len, insertBefore);
        const realPos = col + offset;
        const text = this.text;
        this.text = text.substring(0, realPos) + str + text.substring(realPos);
        return this;
    ***REMOVED***,

    findOffset(pos, len, insertBefore) {
        const offsets = this.offsets;
        let offsetObj;
        let cumulativeOffset = 0;
        let i;

        for (i = 0; i < offsets.length; i += 1) {
            offsetObj = offsets[i];
            if (
                offsetObj.pos < pos ||
                (offsetObj.pos === pos && !insertBefore)
            ) {
                cumulativeOffset += offsetObj.len;
            ***REMOVED***
            if (offsetObj.pos >= pos) {
                break;
            ***REMOVED***
        ***REMOVED***
        if (offsetObj && offsetObj.pos === pos) {
            offsetObj.len += len;
        ***REMOVED*** else {
            offsets.splice(i, 0, { pos, len ***REMOVED***);
        ***REMOVED***
        return cumulativeOffset;
    ***REMOVED***,

    wrap(startPos, startText, endPos, endText, consumeBlanks) {
        this.insertAt(startPos, startText, true, consumeBlanks);
        this.insertAt(endPos, endText, false, consumeBlanks);
        return this;
    ***REMOVED***,

    wrapLine(startText, endText) {
        this.wrap(0, startText, this.originalLength(), endText);
    ***REMOVED***,

    toString() {
        return this.text;
    ***REMOVED***
***REMOVED***;

module.exports = InsertionText;
