// pretty-error/lib/defaultStyle.js
// Generated by CoffeeScript 1.8.0
module.exports = function() {
  return {
    'pretty-error': {
      display: 'block',
      marginLeft: '2'
    ***REMOVED***,
    'pretty-error > header': {
      display: 'block'
    ***REMOVED***,
    'pretty-error > header > title > kind': {
      background: 'red',
      color: 'bright-white'
    ***REMOVED***,
    'pretty-error > header > title > wrapper': {
      marginRight: '1',
      color: 'grey'
    ***REMOVED***,
    'pretty-error > header > colon': {
      color: 'grey',
      marginRight: 1
    ***REMOVED***,
    'pretty-error > header > message': {
      color: 'bright-white'
    ***REMOVED***,
    'pretty-error > trace': {
      display: 'block',
      marginTop: 1
    ***REMOVED***,
    'pretty-error > trace > item': {
      display: 'block',
      marginBottom: 1,
      marginLeft: 2,
      bullet: '"<grey>-</grey>"'
    ***REMOVED***,
    'pretty-error > trace > item > header': {
      display: 'block'
    ***REMOVED***,
    'pretty-error > trace > item > header > pointer > file': {
      color: 'bright-yellow'
    ***REMOVED***,
    'pretty-error > trace > item > header > pointer > colon': {
      color: 'grey'
    ***REMOVED***,
    'pretty-error > trace > item > header > pointer > line': {
      color: 'bright-yellow',
      marginRight: 1
    ***REMOVED***,
    'pretty-error > trace > item > header > what': {
      color: 'white'
    ***REMOVED***,
    'pretty-error > trace > item > footer': {
      display: 'block'
    ***REMOVED***,
    'pretty-error > trace > item > footer > addr': {
      display: 'block',
      color: 'grey'
    ***REMOVED***,
    'pretty-error > trace > item > footer > extra': {
      display: 'block',
      color: 'grey'
    ***REMOVED***
  ***REMOVED***;
***REMOVED***;
