// jsdom/lib/jsdom/living/named-properties-window.js
"use strict";
const hasOwnProp = Object.prototype.hasOwnProperty;
const namedPropertiesTracker = require("../named-properties-tracker");
const NODE_TYPE = require("./node-type");
const HTMLCollection = require("./generated/HTMLCollection");
const { treeOrderSorter ***REMOVED*** = require("../utils");
const idlUtils = require("./generated/utils");

function isNamedPropertyElement(element) {
  // (for the name attribute)

  // use hasOwnProperty to make sure contentWindow comes from the prototype,
  // and is not set directly on the node by a script.
  if ("contentWindow" in element && !hasOwnProp.call(element, "contentWindow")) {
    return true;
  ***REMOVED***

  switch (element.nodeName) {
    case "A":
    case "AREA":
    case "EMBED":
    case "FORM":
    case "FRAMESET":
    case "IMG":
    case "OBJECT":
      return true;
    default:
      return false;
  ***REMOVED***
***REMOVED***

function namedPropertyResolver(window, name, values) {
  function getResult() {
    const results = [];

    for (const node of values().keys()) {
      if (node.nodeType !== NODE_TYPE.ELEMENT_NODE) {
        continue;
      ***REMOVED***

      if (node.getAttributeNS(null, "id") === name) {
        results.push(node);
      ***REMOVED*** else if (node.getAttributeNS(null, "name") === name && isNamedPropertyElement(node)) {
        results.push(node);
      ***REMOVED***
    ***REMOVED***

    results.sort(treeOrderSorter);

    return results;
  ***REMOVED***

  const document = window._document;
  const objects = HTMLCollection.create(window, [], {
    element: idlUtils.implForWrapper(document.documentElement),
    query: getResult
  ***REMOVED***);

  const { length ***REMOVED*** = objects;
  for (let i = 0; i < length; ++i) {
    const node = objects[i];

    if ("contentWindow" in node && !hasOwnProp.call(node, "contentWindow") &&
       node.getAttributeNS(null, "name") === name) {
      return node.contentWindow;
    ***REMOVED***
  ***REMOVED***

  if (length === 0) {
    return undefined;
  ***REMOVED***

  if (length === 1) {
    return objects[0];
  ***REMOVED***

  return objects;
***REMOVED***

exports.initializeWindow = function (window, windowProxy) {
  namedPropertiesTracker.create(window, windowProxy, namedPropertyResolver.bind(null));
***REMOVED***;

exports.elementAttributeModified = function (element, name, value, oldValue) {
  if (!element._attached) {
    return;
  ***REMOVED***

  const useName = isNamedPropertyElement(element);

  if (name === "id" || (name === "name" && useName)) {
    const tracker = namedPropertiesTracker.get(element._ownerDocument._global);

    // (tracker will be null if the document has no Window)
    if (tracker) {
      if (name === "id" && (!useName || element.getAttributeNS(null, "name") !== oldValue)) {
        tracker.untrack(oldValue, element);
      ***REMOVED***

      if (name === "name" && element.getAttributeNS(null, "id") !== oldValue) {
        tracker.untrack(oldValue, element);
      ***REMOVED***

      tracker.track(value, element);
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

exports.nodeAttachedToDocument = function (node) {
  if (node.nodeType !== NODE_TYPE.ELEMENT_NODE) {
    return;
  ***REMOVED***

  const tracker = namedPropertiesTracker.get(node._ownerDocument._global);
  if (!tracker) {
    return;
  ***REMOVED***

  tracker.track(node.getAttributeNS(null, "id"), node);

  if (isNamedPropertyElement(node)) {
    tracker.track(node.getAttributeNS(null, "name"), node);
  ***REMOVED***
***REMOVED***;

exports.nodeDetachedFromDocument = function (node) {
  if (node.nodeType !== NODE_TYPE.ELEMENT_NODE) {
    return;
  ***REMOVED***

  const tracker = namedPropertiesTracker.get(node._ownerDocument._global);
  if (!tracker) {
    return;
  ***REMOVED***

  tracker.untrack(node.getAttributeNS(null, "id"), node);

  if (isNamedPropertyElement(node)) {
    tracker.untrack(node.getAttributeNS(null, "name"), node);
  ***REMOVED***
***REMOVED***;
