// @webassemblyjs/ast/esm/node-path.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; ***REMOVED***); ***REMOVED*** keys.push.apply(keys, symbols); ***REMOVED*** return keys; ***REMOVED***

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {***REMOVED***; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); ***REMOVED***); ***REMOVED*** else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); ***REMOVED*** else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); ***REMOVED***); ***REMOVED*** ***REMOVED*** return target; ***REMOVED***

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true ***REMOVED***); ***REMOVED*** else { obj[key] = value; ***REMOVED*** return obj; ***REMOVED***

function findParent(_ref, cb) {
  var parentPath = _ref.parentPath;

  if (parentPath == null) {
    throw new Error("node is root");
  ***REMOVED***

  var currentPath = parentPath;

  while (cb(currentPath) !== false) {
    // Hit the root node, stop
    // $FlowIgnore
    if (currentPath.parentPath == null) {
      return null;
    ***REMOVED*** // $FlowIgnore


    currentPath = currentPath.parentPath;
  ***REMOVED***

  return currentPath.node;
***REMOVED***

function insertBefore(context, newNode) {
  return insert(context, newNode);
***REMOVED***

function insertAfter(context, newNode) {
  return insert(context, newNode, 1);
***REMOVED***

function insert(_ref2, newNode) {
  var node = _ref2.node,
      inList = _ref2.inList,
      parentPath = _ref2.parentPath,
      parentKey = _ref2.parentKey;
  var indexOffset = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;

  if (!inList) {
    throw new Error('inList' + " error: " + ("insert can only be used for nodes that are within lists" || "unknown"));
  ***REMOVED***

  if (!(parentPath != null)) {
    throw new Error('parentPath != null' + " error: " + ("Can not remove root node" || "unknown"));
  ***REMOVED***

  // $FlowIgnore
  var parentList = parentPath.node[parentKey];
  var indexInList = parentList.findIndex(function (n) {
    return n === node;
  ***REMOVED***);
  parentList.splice(indexInList + indexOffset, 0, newNode);
***REMOVED***

function remove(_ref3) {
  var node = _ref3.node,
      parentKey = _ref3.parentKey,
      parentPath = _ref3.parentPath;

  if (!(parentPath != null)) {
    throw new Error('parentPath != null' + " error: " + ("Can not remove root node" || "unknown"));
  ***REMOVED***

  // $FlowIgnore
  var parentNode = parentPath.node; // $FlowIgnore

  var parentProperty = parentNode[parentKey];

  if (Array.isArray(parentProperty)) {
    // $FlowIgnore
    parentNode[parentKey] = parentProperty.filter(function (n) {
      return n !== node;
    ***REMOVED***);
  ***REMOVED*** else {
    // $FlowIgnore
    delete parentNode[parentKey];
  ***REMOVED***

  node._deleted = true;
***REMOVED***

function stop(context) {
  context.shouldStop = true;
***REMOVED***

function replaceWith(context, newNode) {
  // $FlowIgnore
  var parentNode = context.parentPath.node; // $FlowIgnore

  var parentProperty = parentNode[context.parentKey];

  if (Array.isArray(parentProperty)) {
    var indexInList = parentProperty.findIndex(function (n) {
      return n === context.node;
    ***REMOVED***);
    parentProperty.splice(indexInList, 1, newNode);
  ***REMOVED*** else {
    // $FlowIgnore
    parentNode[context.parentKey] = newNode;
  ***REMOVED***

  context.node._deleted = true;
  context.node = newNode;
***REMOVED*** // bind the context to the first argument of node operations


function bindNodeOperations(operations, context) {
  var keys = Object.keys(operations);
  var boundOperations = {***REMOVED***;
  keys.forEach(function (key) {
    boundOperations[key] = operations[key].bind(null, context);
  ***REMOVED***);
  return boundOperations;
***REMOVED***

function createPathOperations(context) {
  // $FlowIgnore
  return bindNodeOperations({
    findParent: findParent,
    replaceWith: replaceWith,
    remove: remove,
    insertBefore: insertBefore,
    insertAfter: insertAfter,
    stop: stop
  ***REMOVED***, context);
***REMOVED***

export function createPath(context) {
  var path = _objectSpread({***REMOVED***, context); // $FlowIgnore


  Object.assign(path, createPathOperations(path)); // $FlowIgnore

  return path;
***REMOVED***
