// @webassemblyjs/ast/esm/node-helpers.js
import { parse32F, parse64F, parse32I, parse64I, parseU32, isNanLiteral, isInfLiteral ***REMOVED*** from "@webassemblyjs/helper-numbers";
import { longNumberLiteral, floatLiteral, numberLiteral, instr ***REMOVED*** from "./nodes";
export function numberLiteralFromRaw(rawValue) {
  var instructionType = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "i32";
  var original = rawValue; // Remove numeric separators _

  if (typeof rawValue === "string") {
    rawValue = rawValue.replace(/_/g, "");
  ***REMOVED***

  if (typeof rawValue === "number") {
    return numberLiteral(rawValue, String(original));
  ***REMOVED*** else {
    switch (instructionType) {
      case "i32":
        {
          return numberLiteral(parse32I(rawValue), String(original));
        ***REMOVED***

      case "u32":
        {
          return numberLiteral(parseU32(rawValue), String(original));
        ***REMOVED***

      case "i64":
        {
          return longNumberLiteral(parse64I(rawValue), String(original));
        ***REMOVED***

      case "f32":
        {
          return floatLiteral(parse32F(rawValue), isNanLiteral(rawValue), isInfLiteral(rawValue), String(original));
        ***REMOVED***
      // f64

      default:
        {
          return floatLiteral(parse64F(rawValue), isNanLiteral(rawValue), isInfLiteral(rawValue), String(original));
        ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***
export function instruction(id) {
  var args = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
  var namedArgs = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {***REMOVED***;
  return instr(id, undefined, args, namedArgs);
***REMOVED***
export function objectInstruction(id, object) {
  var args = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
  var namedArgs = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {***REMOVED***;
  return instr(id, object, args, namedArgs);
***REMOVED***
/**
 * Decorators
 */

export function withLoc(n, end, start) {
  var loc = {
    start: start,
    end: end
  ***REMOVED***;
  n.loc = loc;
  return n;
***REMOVED***
export function withRaw(n, raw) {
  n.raw = raw;
  return n;
***REMOVED***
export function funcParam(valtype, id) {
  return {
    id: id,
    valtype: valtype
  ***REMOVED***;
***REMOVED***
export function indexLiteral(value) {
  // $FlowIgnore
  var x = numberLiteralFromRaw(value, "u32");
  return x;
***REMOVED***
export function memIndexLiteral(value) {
  // $FlowIgnore
  var x = numberLiteralFromRaw(value, "u32");
  return x;
***REMOVED***
