// clean-css/lib/reader/input-source-map-tracker.js
var SourceMapConsumer = require('source-map').SourceMapConsumer;

function inputSourceMapTracker() {
  var maps = {***REMOVED***;

  return {
    all: all.bind(null, maps),
    isTracking: isTracking.bind(null, maps),
    originalPositionFor: originalPositionFor.bind(null, maps),
    track: track.bind(null, maps)
  ***REMOVED***;
***REMOVED***

function all(maps) {
  return maps;
***REMOVED***

function isTracking(maps, source) {
  return source in maps;
***REMOVED***

function originalPositionFor(maps, metadata, range, selectorFallbacks) {
  var line = metadata[0];
  var column = metadata[1];
  var source = metadata[2];
  var position = {
    line: line,
    column: column + range
  ***REMOVED***;
  var originalPosition;

  while (!originalPosition && position.column > column) {
    position.column--;
    originalPosition = maps[source].originalPositionFor(position);
  ***REMOVED***

  if (!originalPosition || originalPosition.column < 0) {
    return metadata;
  ***REMOVED***

  if (originalPosition.line === null && line > 1 && selectorFallbacks > 0) {
    return originalPositionFor(maps, [line - 1, column, source], range, selectorFallbacks - 1);
  ***REMOVED***

  return originalPosition.line !== null
    ? toMetadata(originalPosition)
    : metadata;
***REMOVED***

function toMetadata(asHash) {
  return [asHash.line, asHash.column, asHash.source];
***REMOVED***

function track(maps, source, data) {
  maps[source] = new SourceMapConsumer(data);
***REMOVED***

module.exports = inputSourceMapTracker;
