// fastq/queue.js
'use strict'

/* eslint-disable no-var */

var reusify = require('reusify')

function fastqueue (context, worker, _concurrency) {
  if (typeof context === 'function') {
    _concurrency = worker
    worker = context
    context = null
  ***REMOVED***

  if (!(_concurrency >= 1)) {
    throw new Error('fastqueue concurrency must be equal to or greater than 1')
  ***REMOVED***

  var cache = reusify(Task)
  var queueHead = null
  var queueTail = null
  var _running = 0
  var errorHandler = null

  var self = {
    push: push,
    drain: noop,
    saturated: noop,
    pause: pause,
    paused: false,

    get concurrency () {
      return _concurrency
    ***REMOVED***,
    set concurrency (value) {
      if (!(value >= 1)) {
        throw new Error('fastqueue concurrency must be equal to or greater than 1')
      ***REMOVED***
      _concurrency = value

      if (self.paused) return
      for (; queueHead && _running < _concurrency;) {
        _running++
        release()
      ***REMOVED***
    ***REMOVED***,

    running: running,
    resume: resume,
    idle: idle,
    length: length,
    getQueue: getQueue,
    unshift: unshift,
    empty: noop,
    kill: kill,
    killAndDrain: killAndDrain,
    error: error
  ***REMOVED***

  return self

  function running () {
    return _running
  ***REMOVED***

  function pause () {
    self.paused = true
  ***REMOVED***

  function length () {
    var current = queueHead
    var counter = 0

    while (current) {
      current = current.next
      counter++
    ***REMOVED***

    return counter
  ***REMOVED***

  function getQueue () {
    var current = queueHead
    var tasks = []

    while (current) {
      tasks.push(current.value)
      current = current.next
    ***REMOVED***

    return tasks
  ***REMOVED***

  function resume () {
    if (!self.paused) return
    self.paused = false
    if (queueHead === null) {
      _running++
      release()
      return
    ***REMOVED***
    for (; queueHead && _running < _concurrency;) {
      _running++
      release()
    ***REMOVED***
  ***REMOVED***

  function idle () {
    return _running === 0 && self.length() === 0
  ***REMOVED***

  function push (value, done) {
    var current = cache.get()

    current.context = context
    current.release = release
    current.value = value
    current.callback = done || noop
    current.errorHandler = errorHandler

    if (_running >= _concurrency || self.paused) {
      if (queueTail) {
        queueTail.next = current
        queueTail = current
      ***REMOVED*** else {
        queueHead = current
        queueTail = current
        self.saturated()
      ***REMOVED***
    ***REMOVED*** else {
      _running++
      worker.call(context, current.value, current.worked)
    ***REMOVED***
  ***REMOVED***

  function unshift (value, done) {
    var current = cache.get()

    current.context = context
    current.release = release
    current.value = value
    current.callback = done || noop
    current.errorHandler = errorHandler

    if (_running >= _concurrency || self.paused) {
      if (queueHead) {
        current.next = queueHead
        queueHead = current
      ***REMOVED*** else {
        queueHead = current
        queueTail = current
        self.saturated()
      ***REMOVED***
    ***REMOVED*** else {
      _running++
      worker.call(context, current.value, current.worked)
    ***REMOVED***
  ***REMOVED***

  function release (holder) {
    if (holder) {
      cache.release(holder)
    ***REMOVED***
    var next = queueHead
    if (next && _running <= _concurrency) {
      if (!self.paused) {
        if (queueTail === queueHead) {
          queueTail = null
        ***REMOVED***
        queueHead = next.next
        next.next = null
        worker.call(context, next.value, next.worked)
        if (queueTail === null) {
          self.empty()
        ***REMOVED***
      ***REMOVED*** else {
        _running--
      ***REMOVED***
    ***REMOVED*** else if (--_running === 0) {
      self.drain()
    ***REMOVED***
  ***REMOVED***

  function kill () {
    queueHead = null
    queueTail = null
    self.drain = noop
  ***REMOVED***

  function killAndDrain () {
    queueHead = null
    queueTail = null
    self.drain()
    self.drain = noop
  ***REMOVED***

  function error (handler) {
    errorHandler = handler
  ***REMOVED***
***REMOVED***

function noop () {***REMOVED***

function Task () {
  this.value = null
  this.callback = noop
  this.next = null
  this.release = noop
  this.context = null
  this.errorHandler = null

  var self = this

  this.worked = function worked (err, result) {
    var callback = self.callback
    var errorHandler = self.errorHandler
    var val = self.value
    self.value = null
    self.callback = noop
    if (self.errorHandler) {
      errorHandler(err, val)
    ***REMOVED***
    callback.call(self.context, err, result)
    self.release(self)
  ***REMOVED***
***REMOVED***

function queueAsPromised (context, worker, _concurrency) {
  if (typeof context === 'function') {
    _concurrency = worker
    worker = context
    context = null
  ***REMOVED***

  function asyncWrapper (arg, cb) {
    worker.call(this, arg)
      .then(function (res) {
        cb(null, res)
      ***REMOVED***, cb)
  ***REMOVED***

  var queue = fastqueue(context, asyncWrapper, _concurrency)

  var pushCb = queue.push
  var unshiftCb = queue.unshift

  queue.push = push
  queue.unshift = unshift
  queue.drained = drained

  return queue

  function push (value) {
    var p = new Promise(function (resolve, reject) {
      pushCb(value, function (err, result) {
        if (err) {
          reject(err)
          return
        ***REMOVED***
        resolve(result)
      ***REMOVED***)
    ***REMOVED***)

    // Let's fork the promise chain to
    // make the error bubble up to the user but
    // not lead to a unhandledRejection
    p.catch(noop)

    return p
  ***REMOVED***

  function unshift (value) {
    var p = new Promise(function (resolve, reject) {
      unshiftCb(value, function (err, result) {
        if (err) {
          reject(err)
          return
        ***REMOVED***
        resolve(result)
      ***REMOVED***)
    ***REMOVED***)

    // Let's fork the promise chain to
    // make the error bubble up to the user but
    // not lead to a unhandledRejection
    p.catch(noop)

    return p
  ***REMOVED***

  function drained () {
    var p = new Promise(function (resolve) {
      process.nextTick(function () {
        if (queue.idle()) {
          resolve()
        ***REMOVED*** else {
          var previousDrain = queue.drain
          queue.drain = function () {
            if (typeof previousDrain === 'function') previousDrain()
            resolve()
            queue.drain = previousDrain
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***)
    ***REMOVED***)

    return p
  ***REMOVED***
***REMOVED***

module.exports = fastqueue
module.exports.promise = queueAsPromised
