// webpack/lib/schemes/HttpUriPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const EventEmitter = require("events");
const { extname, basename ***REMOVED*** = require("path");
const { URL ***REMOVED*** = require("url");
const { createGunzip, createBrotliDecompress, createInflate ***REMOVED*** = require("zlib");
const NormalModule = require("../NormalModule");
const createSchemaValidation = require("../util/create-schema-validation");
const createHash = require("../util/createHash");
const { mkdirp, dirname, join ***REMOVED*** = require("../util/fs");
const memoize = require("../util/memoize");

/** @typedef {import("http").IncomingMessage***REMOVED*** IncomingMessage */
/** @typedef {import("http").OutgoingHttpHeaders***REMOVED*** OutgoingHttpHeaders */
/** @typedef {import("http").RequestOptions***REMOVED*** RequestOptions */
/** @typedef {import("net").Socket***REMOVED*** Socket */
/** @typedef {import("stream").Readable***REMOVED*** Readable */
/** @typedef {import("../../declarations/plugins/schemes/HttpUriPlugin").HttpUriPluginOptions***REMOVED*** HttpUriPluginOptions */
/** @typedef {import("../Compiler")***REMOVED*** Compiler */
/** @typedef {import("../FileSystemInfo").Snapshot***REMOVED*** Snapshot */
/** @typedef {import("../Module").BuildInfo***REMOVED*** BuildInfo */
/** @typedef {import("../NormalModuleFactory").ResourceDataWithData***REMOVED*** ResourceDataWithData */
/** @typedef {import("../util/fs").IntermediateFileSystem***REMOVED*** IntermediateFileSystem */

const getHttp = memoize(() => require("http"));
const getHttps = memoize(() => require("https"));

/**
 * @param {typeof import("http") | typeof import("https")***REMOVED*** request request
 * @param {string | { toString: () => string ***REMOVED*** | undefined***REMOVED*** proxy proxy
 * @returns {(url: URL, requestOptions: RequestOptions, callback: (incomingMessage: IncomingMessage) => void) => EventEmitter***REMOVED*** fn
 */
const proxyFetch = (request, proxy) => (url, options, callback) => {
	const eventEmitter = new EventEmitter();

	/**
	 * @param {Socket=***REMOVED*** socket socket
	 * @returns {void***REMOVED***
	 */
	const doRequest = socket => {
		request
			.get(url, { ...options, ...(socket && { socket ***REMOVED***) ***REMOVED***, callback)
			.on("error", eventEmitter.emit.bind(eventEmitter, "error"));
	***REMOVED***;

	if (proxy) {
		const { hostname: host, port ***REMOVED*** = new URL(proxy);

		getHttp()
			.request({
				host, // IP address of proxy server
				port, // port of proxy server
				method: "CONNECT",
				path: url.host
			***REMOVED***)
			.on("connect", (res, socket) => {
				if (res.statusCode === 200) {
					// connected to proxy server
					doRequest(socket);
				***REMOVED***
			***REMOVED***)
			.on("error", err => {
				eventEmitter.emit(
					"error",
					new Error(
						`Failed to connect to proxy server "${proxy***REMOVED***": ${err.message***REMOVED***`
					)
				);
			***REMOVED***)
			.end();
	***REMOVED*** else {
		doRequest();
	***REMOVED***

	return eventEmitter;
***REMOVED***;

/** @typedef {() => void***REMOVED*** InProgressWriteItem */
/** @type {InProgressWriteItem[] | undefined***REMOVED*** */
let inProgressWrite;

const validate = createSchemaValidation(
	require("../../schemas/plugins/schemes/HttpUriPlugin.check.js"),
	() => require("../../schemas/plugins/schemes/HttpUriPlugin.json"),
	{
		name: "Http Uri Plugin",
		baseDataPath: "options"
	***REMOVED***
);

/**
 * @param {string***REMOVED*** str path
 * @returns {string***REMOVED*** safe path
 */
const toSafePath = str =>
	str
		.replace(/^[^a-zA-Z0-9]+|[^a-zA-Z0-9]+$/g, "")
		.replace(/[^a-zA-Z0-9._-]+/g, "_");

/**
 * @param {Buffer***REMOVED*** content content
 * @returns {string***REMOVED*** integrity
 */
const computeIntegrity = content => {
	const hash = createHash("sha512");
	hash.update(content);
	const integrity = `sha512-${hash.digest("base64")***REMOVED***`;
	return integrity;
***REMOVED***;

/**
 * @param {Buffer***REMOVED*** content content
 * @param {string***REMOVED*** integrity integrity
 * @returns {boolean***REMOVED*** true, if integrity matches
 */
const verifyIntegrity = (content, integrity) => {
	if (integrity === "ignore") return true;
	return computeIntegrity(content) === integrity;
***REMOVED***;

/**
 * @param {string***REMOVED*** str input
 * @returns {Record<string, string>***REMOVED*** parsed
 */
const parseKeyValuePairs = str => {
	/** @type {Record<string, string>***REMOVED*** */
	const result = {***REMOVED***;
	for (const item of str.split(",")) {
		const i = item.indexOf("=");
		if (i >= 0) {
			const key = item.slice(0, i).trim();
			const value = item.slice(i + 1).trim();
			result[key] = value;
		***REMOVED*** else {
			const key = item.trim();
			if (!key) continue;
			result[key] = key;
		***REMOVED***
	***REMOVED***
	return result;
***REMOVED***;

/**
 * @param {string | undefined***REMOVED*** cacheControl Cache-Control header
 * @param {number***REMOVED*** requestTime timestamp of request
 * @returns {{ storeCache: boolean, storeLock: boolean, validUntil: number ***REMOVED******REMOVED*** Logic for storing in cache and lockfile cache
 */
const parseCacheControl = (cacheControl, requestTime) => {
	// When false resource is not stored in cache
	let storeCache = true;
	// When false resource is not stored in lockfile cache
	let storeLock = true;
	// Resource is only revalidated, after that timestamp and when upgrade is chosen
	let validUntil = 0;
	if (cacheControl) {
		const parsed = parseKeyValuePairs(cacheControl);
		if (parsed["no-cache"]) storeCache = storeLock = false;
		if (parsed["max-age"] && !Number.isNaN(Number(parsed["max-age"]))) {
			validUntil = requestTime + Number(parsed["max-age"]) * 1000;
		***REMOVED***
		if (parsed["must-revalidate"]) validUntil = 0;
	***REMOVED***
	return {
		storeLock,
		storeCache,
		validUntil
	***REMOVED***;
***REMOVED***;

/**
 * @typedef {object***REMOVED*** LockfileEntry
 * @property {string***REMOVED*** resolved
 * @property {string***REMOVED*** integrity
 * @property {string***REMOVED*** contentType
 */

/**
 * @param {LockfileEntry***REMOVED*** a first lockfile entry
 * @param {LockfileEntry***REMOVED*** b second lockfile entry
 * @returns {boolean***REMOVED*** true when equal, otherwise false
 */
const areLockfileEntriesEqual = (a, b) =>
	a.resolved === b.resolved &&
	a.integrity === b.integrity &&
	a.contentType === b.contentType;

/**
 * @param {LockfileEntry***REMOVED*** entry lockfile entry
 * @returns {`resolved: ${string***REMOVED***, integrity: ${string***REMOVED***, contentType: ${string***REMOVED***`***REMOVED*** stringified entry
 */
const entryToString = entry =>
	`resolved: ${entry.resolved***REMOVED***, integrity: ${entry.integrity***REMOVED***, contentType: ${entry.contentType***REMOVED***`;

class Lockfile {
	constructor() {
		this.version = 1;
		/** @type {Map<string, LockfileEntry | "ignore" | "no-cache">***REMOVED*** */
		this.entries = new Map();
	***REMOVED***

	/**
	 * @param {string***REMOVED*** content content of the lockfile
	 * @returns {Lockfile***REMOVED*** lockfile
	 */
	static parse(content) {
		// TODO handle merge conflicts
		const data = JSON.parse(content);
		if (data.version !== 1)
			throw new Error(`Unsupported lockfile version ${data.version***REMOVED***`);
		const lockfile = new Lockfile();
		for (const key of Object.keys(data)) {
			if (key === "version") continue;
			const entry = data[key];
			lockfile.entries.set(
				key,
				typeof entry === "string"
					? entry
					: {
							resolved: key,
							...entry
						***REMOVED***
			);
		***REMOVED***
		return lockfile;
	***REMOVED***

	/**
	 * @returns {string***REMOVED*** stringified lockfile
	 */
	toString() {
		let str = "{\n";
		const entries = Array.from(this.entries).sort(([a], [b]) =>
			a < b ? -1 : 1
		);
		for (const [key, entry] of entries) {
			if (typeof entry === "string") {
				str += `  ${JSON.stringify(key)***REMOVED***: ${JSON.stringify(entry)***REMOVED***,\n`;
			***REMOVED*** else {
				str += `  ${JSON.stringify(key)***REMOVED***: { `;
				if (entry.resolved !== key)
					str += `"resolved": ${JSON.stringify(entry.resolved)***REMOVED***, `;
				str += `"integrity": ${JSON.stringify(
					entry.integrity
				)***REMOVED***, "contentType": ${JSON.stringify(entry.contentType)***REMOVED*** ***REMOVED***,\n`;
			***REMOVED***
		***REMOVED***
		str += `  "version": ${this.version***REMOVED***\n***REMOVED***\n`;
		return str;
	***REMOVED***
***REMOVED***

/**
 * @template R
 * @typedef {(err: Error | null, result?: R) => void***REMOVED***  FnWithoutKeyCallback
 */

/**
 * @template R
 * @typedef {(callback: FnWithoutKeyCallback<R>) => void***REMOVED*** FnWithoutKey
 */

/**
 * @template R
 * @param {FnWithoutKey<R>***REMOVED*** fn function
 * @returns {FnWithoutKey<R>***REMOVED*** cached function
 */
const cachedWithoutKey = fn => {
	let inFlight = false;
	/** @type {Error | undefined***REMOVED*** */
	let cachedError;
	/** @type {R | undefined***REMOVED*** */
	let cachedResult;
	/** @type {FnWithoutKeyCallback<R>[] | undefined***REMOVED*** */
	let cachedCallbacks;
	return callback => {
		if (inFlight) {
			if (cachedResult !== undefined) return callback(null, cachedResult);
			if (cachedError !== undefined) return callback(cachedError);
			if (cachedCallbacks === undefined) cachedCallbacks = [callback];
			else cachedCallbacks.push(callback);
			return;
		***REMOVED***
		inFlight = true;
		fn((err, result) => {
			if (err) cachedError = err;
			else cachedResult = result;
			const callbacks = cachedCallbacks;
			cachedCallbacks = undefined;
			callback(err, result);
			if (callbacks !== undefined) for (const cb of callbacks) cb(err, result);
		***REMOVED***);
	***REMOVED***;
***REMOVED***;

/**
 * @template R
 * @typedef {(err: Error | null, result?: R) => void***REMOVED*** FnWithKeyCallback
 */

/**
 * @template T
 * @template R
 * @typedef {(item: T, callback: FnWithKeyCallback<R>) => void***REMOVED*** FnWithKey
 */

/**
 * @template T
 * @template R
 * @param {FnWithKey<T, R>***REMOVED*** fn function
 * @param {FnWithKey<T, R>=***REMOVED*** forceFn function for the second try
 * @returns {(FnWithKey<T, R>) & { force: FnWithKey<T, R> ***REMOVED******REMOVED*** cached function
 */
const cachedWithKey = (fn, forceFn = fn) => {
	/**
	 * @template R
	 * @typedef {{ result?: R, error?: Error, callbacks?: FnWithKeyCallback<R>[], force?: true ***REMOVED******REMOVED*** CacheEntry
	 */
	/** @type {Map<T, CacheEntry<R>>***REMOVED*** */
	const cache = new Map();
	/**
	 * @param {T***REMOVED*** arg arg
	 * @param {FnWithKeyCallback<R>***REMOVED*** callback callback
	 * @returns {void***REMOVED***
	 */
	const resultFn = (arg, callback) => {
		const cacheEntry = cache.get(arg);
		if (cacheEntry !== undefined) {
			if (cacheEntry.result !== undefined)
				return callback(null, cacheEntry.result);
			if (cacheEntry.error !== undefined) return callback(cacheEntry.error);
			if (cacheEntry.callbacks === undefined) cacheEntry.callbacks = [callback];
			else cacheEntry.callbacks.push(callback);
			return;
		***REMOVED***
		/** @type {CacheEntry<R>***REMOVED*** */
		const newCacheEntry = {
			result: undefined,
			error: undefined,
			callbacks: undefined
		***REMOVED***;
		cache.set(arg, newCacheEntry);
		fn(arg, (err, result) => {
			if (err) newCacheEntry.error = err;
			else newCacheEntry.result = result;
			const callbacks = newCacheEntry.callbacks;
			newCacheEntry.callbacks = undefined;
			callback(err, result);
			if (callbacks !== undefined) for (const cb of callbacks) cb(err, result);
		***REMOVED***);
	***REMOVED***;
	/**
	 * @param {T***REMOVED*** arg arg
	 * @param {FnWithKeyCallback<R>***REMOVED*** callback callback
	 * @returns {void***REMOVED***
	 */
	resultFn.force = (arg, callback) => {
		const cacheEntry = cache.get(arg);
		if (cacheEntry !== undefined && cacheEntry.force) {
			if (cacheEntry.result !== undefined)
				return callback(null, cacheEntry.result);
			if (cacheEntry.error !== undefined) return callback(cacheEntry.error);
			if (cacheEntry.callbacks === undefined) cacheEntry.callbacks = [callback];
			else cacheEntry.callbacks.push(callback);
			return;
		***REMOVED***
		/** @type {CacheEntry<R>***REMOVED*** */
		const newCacheEntry = {
			result: undefined,
			error: undefined,
			callbacks: undefined,
			force: true
		***REMOVED***;
		cache.set(arg, newCacheEntry);
		forceFn(arg, (err, result) => {
			if (err) newCacheEntry.error = err;
			else newCacheEntry.result = result;
			const callbacks = newCacheEntry.callbacks;
			newCacheEntry.callbacks = undefined;
			callback(err, result);
			if (callbacks !== undefined) for (const cb of callbacks) cb(err, result);
		***REMOVED***);
	***REMOVED***;
	return resultFn;
***REMOVED***;

/**
 * @typedef {object***REMOVED*** LockfileCache
 * @property {Lockfile***REMOVED*** lockfile lockfile
 * @property {Snapshot***REMOVED*** snapshot snapshot
 */

/**
 * @typedef {object***REMOVED*** ResolveContentResult
 * @property {LockfileEntry***REMOVED*** entry lockfile entry
 * @property {Buffer***REMOVED*** content content
 * @property {boolean***REMOVED*** storeLock need store lockfile
 */

/** @typedef {{ storeCache: boolean, storeLock: boolean, validUntil: number, etag: string | undefined, fresh: boolean ***REMOVED******REMOVED*** FetchResultMeta */
/** @typedef {FetchResultMeta & { location: string ***REMOVED******REMOVED*** RedirectFetchResult */
/** @typedef {FetchResultMeta & { entry: LockfileEntry, content: Buffer ***REMOVED******REMOVED*** ContentFetchResult */
/** @typedef {RedirectFetchResult | ContentFetchResult***REMOVED*** FetchResult */

const PLUGIN_NAME = "HttpUriPlugin";

class HttpUriPlugin {
	/**
	 * @param {HttpUriPluginOptions***REMOVED*** options options
	 */
	constructor(options) {
		validate(options);
		this._lockfileLocation = options.lockfileLocation;
		this._cacheLocation = options.cacheLocation;
		this._upgrade = options.upgrade;
		this._frozen = options.frozen;
		this._allowedUris = options.allowedUris;
		this._proxy = options.proxy;
	***REMOVED***

	/**
	 * Apply the plugin
	 * @param {Compiler***REMOVED*** compiler the compiler instance
	 * @returns {void***REMOVED***
	 */
	apply(compiler) {
		const proxy =
			this._proxy || process.env.http_proxy || process.env.HTTP_PROXY;
		const schemes = [
			{
				scheme: "http",
				fetch: proxyFetch(getHttp(), proxy)
			***REMOVED***,
			{
				scheme: "https",
				fetch: proxyFetch(getHttps(), proxy)
			***REMOVED***
		];
		/** @type {LockfileCache***REMOVED*** */
		let lockfileCache;
		compiler.hooks.compilation.tap(
			PLUGIN_NAME,
			(compilation, { normalModuleFactory ***REMOVED***) => {
				const intermediateFs =
					/** @type {IntermediateFileSystem***REMOVED*** */
					(compiler.intermediateFileSystem);
				const fs = compilation.inputFileSystem;
				const cache = compilation.getCache(`webpack.${PLUGIN_NAME***REMOVED***`);
				const logger = compilation.getLogger(`webpack.${PLUGIN_NAME***REMOVED***`);
				/** @type {string***REMOVED*** */
				const lockfileLocation =
					this._lockfileLocation ||
					join(
						intermediateFs,
						compiler.context,
						compiler.name
							? `${toSafePath(compiler.name)***REMOVED***.webpack.lock`
							: "webpack.lock"
					);
				/** @type {string | false***REMOVED*** */
				const cacheLocation =
					this._cacheLocation !== undefined
						? this._cacheLocation
						: `${lockfileLocation***REMOVED***.data`;
				const upgrade = this._upgrade || false;
				const frozen = this._frozen || false;
				const hashFunction = "sha512";
				const hashDigest = "hex";
				const hashDigestLength = 20;
				const allowedUris = this._allowedUris;

				let warnedAboutEol = false;

				/** @type {Map<string, string>***REMOVED*** */
				const cacheKeyCache = new Map();
				/**
				 * @param {string***REMOVED*** url the url
				 * @returns {string***REMOVED*** the key
				 */
				const getCacheKey = url => {
					const cachedResult = cacheKeyCache.get(url);
					if (cachedResult !== undefined) return cachedResult;
					const result = _getCacheKey(url);
					cacheKeyCache.set(url, result);
					return result;
				***REMOVED***;

				/**
				 * @param {string***REMOVED*** url the url
				 * @returns {string***REMOVED*** the key
				 */
				const _getCacheKey = url => {
					const parsedUrl = new URL(url);
					const folder = toSafePath(parsedUrl.origin);
					const name = toSafePath(parsedUrl.pathname);
					const query = toSafePath(parsedUrl.search);
					let ext = extname(name);
					if (ext.length > 20) ext = "";
					const basename = ext ? name.slice(0, -ext.length) : name;
					const hash = createHash(hashFunction);
					hash.update(url);
					const digest = hash.digest(hashDigest).slice(0, hashDigestLength);
					return `${folder.slice(-50)***REMOVED***/${`${basename***REMOVED***${
						query ? `_${query***REMOVED***` : ""
					***REMOVED***`.slice(0, 150)***REMOVED***_${digest***REMOVED***${ext***REMOVED***`;
				***REMOVED***;

				const getLockfile = cachedWithoutKey(
					/**
					 * @param {(err: Error | null, lockfile?: Lockfile) => void***REMOVED*** callback callback
					 * @returns {void***REMOVED***
					 */
					callback => {
						const readLockfile = () => {
							intermediateFs.readFile(lockfileLocation, (err, buffer) => {
								if (err && err.code !== "ENOENT") {
									compilation.missingDependencies.add(lockfileLocation);
									return callback(err);
								***REMOVED***
								compilation.fileDependencies.add(lockfileLocation);
								compilation.fileSystemInfo.createSnapshot(
									compiler.fsStartTime,
									buffer ? [lockfileLocation] : [],
									[],
									buffer ? [] : [lockfileLocation],
									{ timestamp: true ***REMOVED***,
									(err, s) => {
										if (err) return callback(err);
										const lockfile = buffer
											? Lockfile.parse(buffer.toString("utf-8"))
											: new Lockfile();
										lockfileCache = {
											lockfile,
											snapshot: /** @type {Snapshot***REMOVED*** */ (s)
										***REMOVED***;
										callback(null, lockfile);
									***REMOVED***
								);
							***REMOVED***);
						***REMOVED***;
						if (lockfileCache) {
							compilation.fileSystemInfo.checkSnapshotValid(
								lockfileCache.snapshot,
								(err, valid) => {
									if (err) return callback(err);
									if (!valid) return readLockfile();
									callback(null, lockfileCache.lockfile);
								***REMOVED***
							);
						***REMOVED*** else {
							readLockfile();
						***REMOVED***
					***REMOVED***
				);

				/** @typedef {Map<string, LockfileEntry | "ignore" | "no-cache">***REMOVED*** LockfileUpdates */

				/** @type {LockfileUpdates | undefined***REMOVED*** */
				let lockfileUpdates;

				/**
				 * @param {Lockfile***REMOVED*** lockfile lockfile instance
				 * @param {string***REMOVED*** url url to store
				 * @param {LockfileEntry | "ignore" | "no-cache"***REMOVED*** entry lockfile entry
				 */
				const storeLockEntry = (lockfile, url, entry) => {
					const oldEntry = lockfile.entries.get(url);
					if (lockfileUpdates === undefined) lockfileUpdates = new Map();
					lockfileUpdates.set(url, entry);
					lockfile.entries.set(url, entry);
					if (!oldEntry) {
						logger.log(`${url***REMOVED*** added to lockfile`);
					***REMOVED*** else if (typeof oldEntry === "string") {
						if (typeof entry === "string") {
							logger.log(`${url***REMOVED*** updated in lockfile: ${oldEntry***REMOVED*** -> ${entry***REMOVED***`);
						***REMOVED*** else {
							logger.log(
								`${url***REMOVED*** updated in lockfile: ${oldEntry***REMOVED*** -> ${entry.resolved***REMOVED***`
							);
						***REMOVED***
					***REMOVED*** else if (typeof entry === "string") {
						logger.log(
							`${url***REMOVED*** updated in lockfile: ${oldEntry.resolved***REMOVED*** -> ${entry***REMOVED***`
						);
					***REMOVED*** else if (oldEntry.resolved !== entry.resolved) {
						logger.log(
							`${url***REMOVED*** updated in lockfile: ${oldEntry.resolved***REMOVED*** -> ${entry.resolved***REMOVED***`
						);
					***REMOVED*** else if (oldEntry.integrity !== entry.integrity) {
						logger.log(`${url***REMOVED*** updated in lockfile: content changed`);
					***REMOVED*** else if (oldEntry.contentType !== entry.contentType) {
						logger.log(
							`${url***REMOVED*** updated in lockfile: ${oldEntry.contentType***REMOVED*** -> ${entry.contentType***REMOVED***`
						);
					***REMOVED*** else {
						logger.log(`${url***REMOVED*** updated in lockfile`);
					***REMOVED***
				***REMOVED***;

				/**
				 * @param {Lockfile***REMOVED*** lockfile lockfile
				 * @param {string***REMOVED*** url url
				 * @param {ResolveContentResult***REMOVED*** result result
				 * @param {(err: Error | null, result?: ResolveContentResult) => void***REMOVED*** callback callback
				 * @returns {void***REMOVED***
				 */
				const storeResult = (lockfile, url, result, callback) => {
					if (result.storeLock) {
						storeLockEntry(lockfile, url, result.entry);
						if (!cacheLocation || !result.content)
							return callback(null, result);
						const key = getCacheKey(result.entry.resolved);
						const filePath = join(intermediateFs, cacheLocation, key);
						mkdirp(intermediateFs, dirname(intermediateFs, filePath), err => {
							if (err) return callback(err);
							intermediateFs.writeFile(filePath, result.content, err => {
								if (err) return callback(err);
								callback(null, result);
							***REMOVED***);
						***REMOVED***);
					***REMOVED*** else {
						storeLockEntry(lockfile, url, "no-cache");
						callback(null, result);
					***REMOVED***
				***REMOVED***;

				for (const { scheme, fetch ***REMOVED*** of schemes) {
					/**
					 * @param {string***REMOVED*** url URL
					 * @param {string | null***REMOVED*** integrity integrity
					 * @param {(err: Error | null, resolveContentResult?: ResolveContentResult) => void***REMOVED*** callback callback
					 */
					const resolveContent = (url, integrity, callback) => {
						/**
						 * @param {Error | null***REMOVED*** err error
						 * @param {FetchResult=***REMOVED*** _result fetch result
						 * @returns {void***REMOVED***
						 */
						const handleResult = (err, _result) => {
							if (err) return callback(err);

							const result = /** @type {FetchResult***REMOVED*** */ (_result);

							if ("location" in result) {
								return resolveContent(
									result.location,
									integrity,
									(err, innerResult) => {
										if (err) return callback(err);
										const { entry, content, storeLock ***REMOVED*** =
											/** @type {ResolveContentResult***REMOVED*** */ (innerResult);
										callback(null, {
											entry,
											content,
											storeLock: storeLock && result.storeLock
										***REMOVED***);
									***REMOVED***
								);
							***REMOVED***

							if (
								!result.fresh &&
								integrity &&
								result.entry.integrity !== integrity &&
								!verifyIntegrity(result.content, integrity)
							) {
								return fetchContent.force(url, handleResult);
							***REMOVED***

							return callback(null, {
								entry: result.entry,
								content: result.content,
								storeLock: result.storeLock
							***REMOVED***);
						***REMOVED***;

						fetchContent(url, handleResult);
					***REMOVED***;

					/**
					 * @param {string***REMOVED*** url URL
					 * @param {FetchResult | RedirectFetchResult | undefined***REMOVED*** cachedResult result from cache
					 * @param {(err: Error | null, fetchResult?: FetchResult) => void***REMOVED*** callback callback
					 * @returns {void***REMOVED***
					 */
					const fetchContentRaw = (url, cachedResult, callback) => {
						const requestTime = Date.now();
						/** @type {OutgoingHttpHeaders***REMOVED*** */
						const headers = {
							"accept-encoding": "gzip, deflate, br",
							"user-agent": "webpack"
						***REMOVED***;

						if (cachedResult && cachedResult.etag) {
							headers["if-none-match"] = cachedResult.etag;
						***REMOVED***

						fetch(new URL(url), { headers ***REMOVED***, res => {
							const etag = res.headers.etag;
							const location = res.headers.location;
							const cacheControl = res.headers["cache-control"];
							const { storeLock, storeCache, validUntil ***REMOVED*** = parseCacheControl(
								cacheControl,
								requestTime
							);
							/**
							 * @param {Partial<Pick<FetchResultMeta, "fresh">> & (Pick<RedirectFetchResult, "location"> | Pick<ContentFetchResult, "content" | "entry">)***REMOVED*** partialResult result
							 * @returns {void***REMOVED***
							 */
							const finishWith = partialResult => {
								if ("location" in partialResult) {
									logger.debug(
										`GET ${url***REMOVED*** [${res.statusCode***REMOVED***] -> ${partialResult.location***REMOVED***`
									);
								***REMOVED*** else {
									logger.debug(
										`GET ${url***REMOVED*** [${res.statusCode***REMOVED***] ${Math.ceil(
											partialResult.content.length / 1024
										)***REMOVED*** kB${!storeLock ? " no-cache" : ""***REMOVED***`
									);
								***REMOVED***
								const result = {
									...partialResult,
									fresh: true,
									storeLock,
									storeCache,
									validUntil,
									etag
								***REMOVED***;
								if (!storeCache) {
									logger.log(
										`${url***REMOVED*** can't be stored in cache, due to Cache-Control header: ${cacheControl***REMOVED***`
									);
									return callback(null, result);
								***REMOVED***
								cache.store(
									url,
									null,
									{
										...result,
										fresh: false
									***REMOVED***,
									err => {
										if (err) {
											logger.warn(
												`${url***REMOVED*** can't be stored in cache: ${err.message***REMOVED***`
											);
											logger.debug(err.stack);
										***REMOVED***
										callback(null, result);
									***REMOVED***
								);
							***REMOVED***;
							if (res.statusCode === 304) {
								const result = /** @type {FetchResult***REMOVED*** */ (cachedResult);
								if (
									result.validUntil < validUntil ||
									result.storeLock !== storeLock ||
									result.storeCache !== storeCache ||
									result.etag !== etag
								) {
									return finishWith(result);
								***REMOVED***
								logger.debug(`GET ${url***REMOVED*** [${res.statusCode***REMOVED***] (unchanged)`);
								return callback(null, { ...result, fresh: true ***REMOVED***);
							***REMOVED***
							if (
								location &&
								res.statusCode &&
								res.statusCode >= 301 &&
								res.statusCode <= 308
							) {
								const result = {
									location: new URL(location, url).href
								***REMOVED***;
								if (
									!cachedResult ||
									!("location" in cachedResult) ||
									cachedResult.location !== result.location ||
									cachedResult.validUntil < validUntil ||
									cachedResult.storeLock !== storeLock ||
									cachedResult.storeCache !== storeCache ||
									cachedResult.etag !== etag
								) {
									return finishWith(result);
								***REMOVED***
								logger.debug(`GET ${url***REMOVED*** [${res.statusCode***REMOVED***] (unchanged)`);
								return callback(null, {
									...result,
									fresh: true,
									storeLock,
									storeCache,
									validUntil,
									etag
								***REMOVED***);
							***REMOVED***
							const contentType = res.headers["content-type"] || "";
							/** @type {Buffer[]***REMOVED*** */
							const bufferArr = [];

							const contentEncoding = res.headers["content-encoding"];
							/** @type {Readable***REMOVED*** */
							let stream = res;
							if (contentEncoding === "gzip") {
								stream = stream.pipe(createGunzip());
							***REMOVED*** else if (contentEncoding === "br") {
								stream = stream.pipe(createBrotliDecompress());
							***REMOVED*** else if (contentEncoding === "deflate") {
								stream = stream.pipe(createInflate());
							***REMOVED***

							stream.on("data", chunk => {
								bufferArr.push(chunk);
							***REMOVED***);

							stream.on("end", () => {
								if (!res.complete) {
									logger.log(`GET ${url***REMOVED*** [${res.statusCode***REMOVED***] (terminated)`);
									return callback(new Error(`${url***REMOVED*** request was terminated`));
								***REMOVED***

								const content = Buffer.concat(bufferArr);

								if (res.statusCode !== 200) {
									logger.log(`GET ${url***REMOVED*** [${res.statusCode***REMOVED***]`);
									return callback(
										new Error(
											`${url***REMOVED*** request status code = ${
												res.statusCode
											***REMOVED***\n${content.toString("utf-8")***REMOVED***`
										)
									);
								***REMOVED***

								const integrity = computeIntegrity(content);
								const entry = { resolved: url, integrity, contentType ***REMOVED***;

								finishWith({
									entry,
									content
								***REMOVED***);
							***REMOVED***);
						***REMOVED***).on("error", err => {
							logger.log(`GET ${url***REMOVED*** (error)`);
							err.message += `\nwhile fetching ${url***REMOVED***`;
							callback(err);
						***REMOVED***);
					***REMOVED***;

					const fetchContent = cachedWithKey(
						/**
						 * @param {string***REMOVED*** url URL
						 * @param {(err: Error | null, result?: FetchResult) => void***REMOVED*** callback callback
						 * @returns {void***REMOVED***
						 */
						(url, callback) => {
							cache.get(url, null, (err, cachedResult) => {
								if (err) return callback(err);
								if (cachedResult) {
									const isValid = cachedResult.validUntil >= Date.now();
									if (isValid) return callback(null, cachedResult);
								***REMOVED***
								fetchContentRaw(url, cachedResult, callback);
							***REMOVED***);
						***REMOVED***,
						(url, callback) => fetchContentRaw(url, undefined, callback)
					);

					/**
					 * @param {string***REMOVED*** uri uri
					 * @returns {boolean***REMOVED*** true when allowed, otherwise false
					 */
					const isAllowed = uri => {
						for (const allowed of allowedUris) {
							if (typeof allowed === "string") {
								if (uri.startsWith(allowed)) return true;
							***REMOVED*** else if (typeof allowed === "function") {
								if (allowed(uri)) return true;
							***REMOVED*** else if (allowed.test(uri)) {
								return true;
							***REMOVED***
						***REMOVED***
						return false;
					***REMOVED***;

					/** @typedef {{ entry: LockfileEntry, content: Buffer ***REMOVED******REMOVED*** Info */

					const getInfo = cachedWithKey(
						/**
						 * @param {string***REMOVED*** url the url
						 * @param {(err: Error | null, info?: Info) => void***REMOVED*** callback callback
						 * @returns {void***REMOVED***
						 */
						// eslint-disable-next-line no-loop-func
						(url, callback) => {
							if (!isAllowed(url)) {
								return callback(
									new Error(
										`${url***REMOVED*** doesn't match the allowedUris policy. These URIs are allowed:\n${allowedUris
											.map(uri => ` - ${uri***REMOVED***`)
											.join("\n")***REMOVED***`
									)
								);
							***REMOVED***
							getLockfile((err, _lockfile) => {
								if (err) return callback(err);
								const lockfile = /** @type {Lockfile***REMOVED*** */ (_lockfile);
								const entryOrString = lockfile.entries.get(url);
								if (!entryOrString) {
									if (frozen) {
										return callback(
											new Error(
												`${url***REMOVED*** has no lockfile entry and lockfile is frozen`
											)
										);
									***REMOVED***
									resolveContent(url, null, (err, result) => {
										if (err) return callback(err);
										storeResult(
											/** @type {Lockfile***REMOVED*** */
											(lockfile),
											url,
											/** @type {ResolveContentResult***REMOVED*** */
											(result),
											callback
										);
									***REMOVED***);
									return;
								***REMOVED***
								if (typeof entryOrString === "string") {
									const entryTag = entryOrString;
									resolveContent(url, null, (err, _result) => {
										if (err) return callback(err);
										const result =
											/** @type {ResolveContentResult***REMOVED*** */
											(_result);
										if (!result.storeLock || entryTag === "ignore")
											return callback(null, result);
										if (frozen) {
											return callback(
												new Error(
													`${url***REMOVED*** used to have ${entryTag***REMOVED*** lockfile entry and has content now, but lockfile is frozen`
												)
											);
										***REMOVED***
										if (!upgrade) {
											return callback(
												new Error(
													`${url***REMOVED*** used to have ${entryTag***REMOVED*** lockfile entry and has content now.
This should be reflected in the lockfile, so this lockfile entry must be upgraded, but upgrading is not enabled.
Remove this line from the lockfile to force upgrading.`
												)
											);
										***REMOVED***
										storeResult(lockfile, url, result, callback);
									***REMOVED***);
									return;
								***REMOVED***
								let entry = entryOrString;
								/**
								 * @param {Buffer=***REMOVED*** lockedContent locked content
								 */
								const doFetch = lockedContent => {
									resolveContent(url, entry.integrity, (err, _result) => {
										if (err) {
											if (lockedContent) {
												logger.warn(
													`Upgrade request to ${url***REMOVED*** failed: ${err.message***REMOVED***`
												);
												logger.debug(err.stack);
												return callback(null, {
													entry,
													content: lockedContent
												***REMOVED***);
											***REMOVED***
											return callback(err);
										***REMOVED***
										const result =
											/** @type {ResolveContentResult***REMOVED*** */
											(_result);
										if (!result.storeLock) {
											// When the lockfile entry should be no-cache
											// we need to update the lockfile
											if (frozen) {
												return callback(
													new Error(
														`${url***REMOVED*** has a lockfile entry and is no-cache now, but lockfile is frozen\nLockfile: ${entryToString(
															entry
														)***REMOVED***`
													)
												);
											***REMOVED***
											storeResult(lockfile, url, result, callback);
											return;
										***REMOVED***
										if (!areLockfileEntriesEqual(result.entry, entry)) {
											// When the lockfile entry is outdated
											// we need to update the lockfile
											if (frozen) {
												return callback(
													new Error(
														`${url***REMOVED*** has an outdated lockfile entry, but lockfile is frozen\nLockfile: ${entryToString(
															entry
														)***REMOVED***\nExpected: ${entryToString(result.entry)***REMOVED***`
													)
												);
											***REMOVED***
											storeResult(lockfile, url, result, callback);
											return;
										***REMOVED***
										if (!lockedContent && cacheLocation) {
											// When the lockfile cache content is missing
											// we need to update the lockfile
											if (frozen) {
												return callback(
													new Error(
														`${url***REMOVED*** is missing content in the lockfile cache, but lockfile is frozen\nLockfile: ${entryToString(
															entry
														)***REMOVED***`
													)
												);
											***REMOVED***
											storeResult(lockfile, url, result, callback);
											return;
										***REMOVED***
										return callback(null, result);
									***REMOVED***);
								***REMOVED***;
								if (cacheLocation) {
									// When there is a lockfile cache
									// we read the content from there
									const key = getCacheKey(entry.resolved);
									const filePath = join(intermediateFs, cacheLocation, key);
									fs.readFile(filePath, (err, result) => {
										if (err) {
											if (err.code === "ENOENT") return doFetch();
											return callback(err);
										***REMOVED***
										const content = /** @type {Buffer***REMOVED*** */ (result);
										/**
										 * @param {Buffer | undefined***REMOVED*** _result result
										 * @returns {void***REMOVED***
										 */
										const continueWithCachedContent = _result => {
											if (!upgrade) {
												// When not in upgrade mode, we accept the result from the lockfile cache
												return callback(null, { entry, content ***REMOVED***);
											***REMOVED***
											return doFetch(content);
										***REMOVED***;
										if (!verifyIntegrity(content, entry.integrity)) {
											/** @type {Buffer | undefined***REMOVED*** */
											let contentWithChangedEol;
											let isEolChanged = false;
											try {
												contentWithChangedEol = Buffer.from(
													content.toString("utf-8").replace(/\r\n/g, "\n")
												);
												isEolChanged = verifyIntegrity(
													contentWithChangedEol,
													entry.integrity
												);
											***REMOVED*** catch (_err) {
												// ignore
											***REMOVED***
											if (isEolChanged) {
												if (!warnedAboutEol) {
													const explainer = `Incorrect end of line sequence was detected in the lockfile cache.
The lockfile cache is protected by integrity checks, so any external modification will lead to a corrupted lockfile cache.
When using git make sure to configure .gitattributes correctly for the lockfile cache:
  **/*webpack.lock.data/** -text
This will avoid that the end of line sequence is changed by git on Windows.`;
													if (frozen) {
														logger.error(explainer);
													***REMOVED*** else {
														logger.warn(explainer);
														logger.info(
															"Lockfile cache will be automatically fixed now, but when lockfile is frozen this would result in an error."
														);
													***REMOVED***
													warnedAboutEol = true;
												***REMOVED***
												if (!frozen) {
													// "fix" the end of line sequence of the lockfile content
													logger.log(
														`${filePath***REMOVED*** fixed end of line sequence (\\r\\n instead of \\n).`
													);
													intermediateFs.writeFile(
														filePath,
														/** @type {Buffer***REMOVED*** */
														(contentWithChangedEol),
														err => {
															if (err) return callback(err);
															continueWithCachedContent(
																/** @type {Buffer***REMOVED*** */
																(contentWithChangedEol)
															);
														***REMOVED***
													);
													return;
												***REMOVED***
											***REMOVED***
											if (frozen) {
												return callback(
													new Error(
														`${
															entry.resolved
														***REMOVED*** integrity mismatch, expected content with integrity ${
															entry.integrity
														***REMOVED*** but got ${computeIntegrity(content)***REMOVED***.
Lockfile corrupted (${
															isEolChanged
																? "end of line sequence was unexpectedly changed"
																: "incorrectly merged? changed by other tools?"
														***REMOVED***).
Run build with un-frozen lockfile to automatically fix lockfile.`
													)
												);
											***REMOVED***
											// "fix" the lockfile entry to the correct integrity
											// the content has priority over the integrity value
											entry = {
												...entry,
												integrity: computeIntegrity(content)
											***REMOVED***;
											storeLockEntry(lockfile, url, entry);
										***REMOVED***
										continueWithCachedContent(result);
									***REMOVED***);
								***REMOVED*** else {
									doFetch();
								***REMOVED***
							***REMOVED***);
						***REMOVED***
					);

					/**
					 * @param {URL***REMOVED*** url url
					 * @param {ResourceDataWithData***REMOVED*** resourceData resource data
					 * @param {(err: Error | null, result: true | void) => void***REMOVED*** callback callback
					 */
					const respondWithUrlModule = (url, resourceData, callback) => {
						getInfo(url.href, (err, _result) => {
							if (err) return callback(err);
							const result = /** @type {Info***REMOVED*** */ (_result);
							resourceData.resource = url.href;
							resourceData.path = url.origin + url.pathname;
							resourceData.query = url.search;
							resourceData.fragment = url.hash;
							resourceData.context = new URL(
								".",
								result.entry.resolved
							).href.slice(0, -1);
							resourceData.data.mimetype = result.entry.contentType;
							callback(null, true);
						***REMOVED***);
					***REMOVED***;
					normalModuleFactory.hooks.resolveForScheme
						.for(scheme)
						.tapAsync(PLUGIN_NAME, (resourceData, resolveData, callback) => {
							respondWithUrlModule(
								new URL(resourceData.resource),
								resourceData,
								callback
							);
						***REMOVED***);
					normalModuleFactory.hooks.resolveInScheme
						.for(scheme)
						.tapAsync(PLUGIN_NAME, (resourceData, data, callback) => {
							// Only handle relative urls (./xxx, ../xxx, /xxx, //xxx)
							if (
								data.dependencyType !== "url" &&
								!/^\.{0,2***REMOVED***\//.test(resourceData.resource)
							) {
								return callback();
							***REMOVED***
							respondWithUrlModule(
								new URL(resourceData.resource, `${data.context***REMOVED***/`),
								resourceData,
								callback
							);
						***REMOVED***);
					const hooks = NormalModule.getCompilationHooks(compilation);
					hooks.readResourceForScheme
						.for(scheme)
						.tapAsync(PLUGIN_NAME, (resource, module, callback) =>
							getInfo(resource, (err, _result) => {
								if (err) return callback(err);
								const result = /** @type {Info***REMOVED*** */ (_result);
								/** @type {BuildInfo***REMOVED*** */
								(module.buildInfo).resourceIntegrity = result.entry.integrity;
								callback(null, result.content);
							***REMOVED***)
						);
					hooks.needBuild.tapAsync(PLUGIN_NAME, (module, context, callback) => {
						if (module.resource && module.resource.startsWith(`${scheme***REMOVED***://`)) {
							getInfo(module.resource, (err, _result) => {
								if (err) return callback(err);
								const result = /** @type {Info***REMOVED*** */ (_result);
								if (
									result.entry.integrity !==
									/** @type {BuildInfo***REMOVED*** */
									(module.buildInfo).resourceIntegrity
								) {
									return callback(null, true);
								***REMOVED***
								callback();
							***REMOVED***);
						***REMOVED*** else {
							return callback();
						***REMOVED***
					***REMOVED***);
				***REMOVED***
				compilation.hooks.finishModules.tapAsync(
					PLUGIN_NAME,
					(modules, callback) => {
						if (!lockfileUpdates) return callback();
						const ext = extname(lockfileLocation);
						const tempFile = join(
							intermediateFs,
							dirname(intermediateFs, lockfileLocation),
							`.${basename(lockfileLocation, ext)***REMOVED***.${
								(Math.random() * 10000) | 0
							***REMOVED***${ext***REMOVED***`
						);

						const writeDone = () => {
							const nextOperation =
								/** @type {InProgressWriteItem[]***REMOVED*** */
								(inProgressWrite).shift();
							if (nextOperation) {
								nextOperation();
							***REMOVED*** else {
								inProgressWrite = undefined;
							***REMOVED***
						***REMOVED***;
						const runWrite = () => {
							intermediateFs.readFile(lockfileLocation, (err, buffer) => {
								if (err && err.code !== "ENOENT") {
									writeDone();
									return callback(err);
								***REMOVED***
								const lockfile = buffer
									? Lockfile.parse(buffer.toString("utf-8"))
									: new Lockfile();
								for (const [key, value] of /** @type {LockfileUpdates***REMOVED*** */ (
									lockfileUpdates
								)) {
									lockfile.entries.set(key, value);
								***REMOVED***
								intermediateFs.writeFile(tempFile, lockfile.toString(), err => {
									if (err) {
										writeDone();
										return (
											/** @type {NonNullable<IntermediateFileSystem["unlink"]>***REMOVED*** */
											(intermediateFs.unlink)(tempFile, () => callback(err))
										);
									***REMOVED***
									intermediateFs.rename(tempFile, lockfileLocation, err => {
										if (err) {
											writeDone();
											return (
												/** @type {NonNullable<IntermediateFileSystem["unlink"]>***REMOVED*** */
												(intermediateFs.unlink)(tempFile, () => callback(err))
											);
										***REMOVED***
										writeDone();
										callback();
									***REMOVED***);
								***REMOVED***);
							***REMOVED***);
						***REMOVED***;
						if (inProgressWrite) {
							inProgressWrite.push(runWrite);
						***REMOVED*** else {
							inProgressWrite = [];
							runWrite();
						***REMOVED***
					***REMOVED***
				);
			***REMOVED***
		);
	***REMOVED***
***REMOVED***

module.exports = HttpUriPlugin;
