// @babel/traverse/lib/scope/binding.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
class Binding {
  constructor({
    identifier,
    scope,
    path,
    kind
  ***REMOVED***) {
    this.identifier = void 0;
    this.scope = void 0;
    this.path = void 0;
    this.kind = void 0;
    this.constantViolations = [];
    this.constant = true;
    this.referencePaths = [];
    this.referenced = false;
    this.references = 0;
    this.identifier = identifier;
    this.scope = scope;
    this.path = path;
    this.kind = kind;
    if ((kind === "var" || kind === "hoisted") && isInitInLoop(path)) {
      this.reassign(path);
    ***REMOVED***
    this.clearValue();
  ***REMOVED***
  deoptValue() {
    this.clearValue();
    this.hasDeoptedValue = true;
  ***REMOVED***
  setValue(value) {
    if (this.hasDeoptedValue) return;
    this.hasValue = true;
    this.value = value;
  ***REMOVED***
  clearValue() {
    this.hasDeoptedValue = false;
    this.hasValue = false;
    this.value = null;
  ***REMOVED***
  reassign(path) {
    this.constant = false;
    if (this.constantViolations.includes(path)) {
      return;
    ***REMOVED***
    this.constantViolations.push(path);
  ***REMOVED***
  reference(path) {
    if (this.referencePaths.includes(path)) {
      return;
    ***REMOVED***
    this.referenced = true;
    this.references++;
    this.referencePaths.push(path);
  ***REMOVED***
  dereference() {
    this.references--;
    this.referenced = !!this.references;
  ***REMOVED***
***REMOVED***
exports.default = Binding;
function isInitInLoop(path) {
  const isFunctionDeclarationOrHasInit = !path.isVariableDeclarator() || path.node.init;
  for (let {
    parentPath,
    key
  ***REMOVED*** = path; parentPath; {
    parentPath,
    key
  ***REMOVED*** = parentPath) {
    if (parentPath.isFunctionParent()) return false;
    if (key === "left" && parentPath.isForXStatement() || isFunctionDeclarationOrHasInit && key === "body" && parentPath.isLoop()) {
      return true;
    ***REMOVED***
  ***REMOVED***
  return false;
***REMOVED***
