// @babel/core/lib/transformation/block-hoist-plugin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = loadBlockHoistPlugin;
function _traverse() {
  const data = require("@babel/traverse");
  _traverse = function () {
    return data;
  ***REMOVED***;
  return data;
***REMOVED***
var _plugin = require("../config/plugin.js");
let LOADED_PLUGIN;
const blockHoistPlugin = {
  name: "internal.blockHoist",
  visitor: {
    Block: {
      exit({
        node
      ***REMOVED***) {
        node.body = performHoisting(node.body);
      ***REMOVED***
    ***REMOVED***,
    SwitchCase: {
      exit({
        node
      ***REMOVED***) {
        node.consequent = performHoisting(node.consequent);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***;
function performHoisting(body) {
  let max = Math.pow(2, 30) - 1;
  let hasChange = false;
  for (let i = 0; i < body.length; i++) {
    const n = body[i];
    const p = priority(n);
    if (p > max) {
      hasChange = true;
      break;
    ***REMOVED***
    max = p;
  ***REMOVED***
  if (!hasChange) return body;
  return stableSort(body.slice());
***REMOVED***
function loadBlockHoistPlugin() {
  if (!LOADED_PLUGIN) {
    LOADED_PLUGIN = new _plugin.default(Object.assign({***REMOVED***, blockHoistPlugin, {
      visitor: _traverse().default.explode(blockHoistPlugin.visitor)
    ***REMOVED***), {***REMOVED***);
  ***REMOVED***
  return LOADED_PLUGIN;
***REMOVED***
function priority(bodyNode) {
  const priority = bodyNode == null ? void 0 : bodyNode._blockHoist;
  if (priority == null) return 1;
  if (priority === true) return 2;
  return priority;
***REMOVED***
function stableSort(body) {
  const buckets = Object.create(null);
  for (let i = 0; i < body.length; i++) {
    const n = body[i];
    const p = priority(n);
    const bucket = buckets[p] || (buckets[p] = []);
    bucket.push(n);
  ***REMOVED***
  const keys = Object.keys(buckets).map(k => +k).sort((a, b) => b - a);
  let index = 0;
  for (const key of keys) {
    const bucket = buckets[key];
    for (const n of bucket) {
      body[index++] = n;
    ***REMOVED***
  ***REMOVED***
  return body;
***REMOVED***
0 && 0;
