// @babel/core/lib/transformation/plugin-pass.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
class PluginPass {
  constructor(file, key, options, isAsync) {
    this._map = new Map();
    this.key = void 0;
    this.file = void 0;
    this.opts = void 0;
    this.cwd = void 0;
    this.filename = void 0;
    this.isAsync = void 0;
    this.key = key;
    this.file = file;
    this.opts = options || {***REMOVED***;
    this.cwd = file.opts.cwd;
    this.filename = file.opts.filename;
    this.isAsync = isAsync;
  ***REMOVED***
  set(key, val) {
    this._map.set(key, val);
  ***REMOVED***
  get(key) {
    return this._map.get(key);
  ***REMOVED***
  availableHelper(name, versionRange) {
    return this.file.availableHelper(name, versionRange);
  ***REMOVED***
  addHelper(name) {
    return this.file.addHelper(name);
  ***REMOVED***
  buildCodeFrameError(node, msg, _Error) {
    return this.file.buildCodeFrameError(node, msg, _Error);
  ***REMOVED***
***REMOVED***
exports.default = PluginPass;
{
  PluginPass.prototype.getModuleName = function getModuleName() {
    return this.file.getModuleName();
  ***REMOVED***;
  PluginPass.prototype.addImport = function addImport() {
    this.file.addImport();
  ***REMOVED***;
***REMOVED***
0 && 0;
