// css-minimizer-webpack-plugin/dist/utils.js
"use strict";

/** @typedef {import("./index.js").Input***REMOVED*** Input */

/** @typedef {import("source-map").RawSourceMap***REMOVED*** RawSourceMap */

/** @typedef {import("source-map").SourceMapGenerator***REMOVED*** SourceMapGenerator */

/** @typedef {import("./index.js").MinimizedResult***REMOVED*** MinimizedResult */

/** @typedef {import("./index.js").CustomOptions***REMOVED*** CustomOptions */

/** @typedef {import("postcss").ProcessOptions***REMOVED*** ProcessOptions */

/** @typedef {import("postcss").Postcss***REMOVED*** Postcss */
const notSettled = Symbol(`not-settled`);
/**
 * @template T
 * @typedef {() => Promise<T>***REMOVED*** Task
 */

/**
 * Run tasks with limited concurency.
 * @template T
 * @param {number***REMOVED*** limit - Limit of tasks that run at once.
 * @param {Task<T>[]***REMOVED*** tasks - List of tasks to run.
 * @returns {Promise<T[]>***REMOVED*** A promise that fulfills to an array of the results
 */

function throttleAll(limit, tasks) {
  if (!Number.isInteger(limit) || limit < 1) {
    throw new TypeError(`Expected \`limit\` to be a finite number > 0, got \`${limit***REMOVED***\` (${typeof limit***REMOVED***)`);
  ***REMOVED***

  if (!Array.isArray(tasks) || !tasks.every(task => typeof task === `function`)) {
    throw new TypeError(`Expected \`tasks\` to be a list of functions returning a promise`);
  ***REMOVED***

  return new Promise((resolve, reject) => {
    const result = Array(tasks.length).fill(notSettled);
    const entries = tasks.entries();

    const next = () => {
      const {
        done,
        value
      ***REMOVED*** = entries.next();

      if (done) {
        const isLast = !result.includes(notSettled);
        if (isLast) resolve(result);
        return;
      ***REMOVED***

      const [index, task] = value;
      /**
       * @param {T***REMOVED*** x
       */

      const onFulfilled = x => {
        result[index] = x;
        next();
      ***REMOVED***;

      task().then(onFulfilled, reject);
    ***REMOVED***;

    Array(limit).fill(0).forEach(next);
  ***REMOVED***);
***REMOVED***
/* istanbul ignore next */

/**
 * @param {Input***REMOVED*** input
 * @param {RawSourceMap | undefined***REMOVED*** sourceMap
 * @param {CustomOptions***REMOVED*** minimizerOptions
 * @return {Promise<MinimizedResult>***REMOVED***
 */


async function cssnanoMinify(input, sourceMap, minimizerOptions = {
  preset: "default"
***REMOVED***) {
  /**
   * @template T
   * @param {string***REMOVED*** module
   * @returns {Promise<T>***REMOVED***
   */
  const load = async module => {
    let exports;

    try {
      // eslint-disable-next-line import/no-dynamic-require, global-require
      exports = require(module);
      return exports;
    ***REMOVED*** catch (requireError) {
      let importESM;

      try {
        // eslint-disable-next-line no-new-func
        importESM = new Function("id", "return import(id);");
      ***REMOVED*** catch (e) {
        importESM = null;
      ***REMOVED***

      if (
      /** @type {Error & {code: string***REMOVED******REMOVED*** */
      requireError.code === "ERR_REQUIRE_ESM" && importESM) {
        exports = await importESM(module);
        return exports.default;
      ***REMOVED***

      throw requireError;
    ***REMOVED***
  ***REMOVED***;

  const [[name, code]] = Object.entries(input);
  /** @type {ProcessOptions***REMOVED*** */

  const postcssOptions = {
    from: name,
    ...minimizerOptions.processorOptions
  ***REMOVED***;

  if (typeof postcssOptions.parser === "string") {
    try {
      postcssOptions.parser = await load(postcssOptions.parser);
    ***REMOVED*** catch (error) {
      throw new Error(`Loading PostCSS "${postcssOptions.parser***REMOVED***" parser failed: ${
      /** @type {Error***REMOVED*** */
      error.message***REMOVED***\n\n(@${name***REMOVED***)`);
    ***REMOVED***
  ***REMOVED***

  if (typeof postcssOptions.stringifier === "string") {
    try {
      postcssOptions.stringifier = await load(postcssOptions.stringifier);
    ***REMOVED*** catch (error) {
      throw new Error(`Loading PostCSS "${postcssOptions.stringifier***REMOVED***" stringifier failed: ${
      /** @type {Error***REMOVED*** */
      error.message***REMOVED***\n\n(@${name***REMOVED***)`);
    ***REMOVED***
  ***REMOVED***

  if (typeof postcssOptions.syntax === "string") {
    try {
      postcssOptions.syntax = await load(postcssOptions.syntax);
    ***REMOVED*** catch (error) {
      throw new Error(`Loading PostCSS "${postcssOptions.syntax***REMOVED***" syntax failed: ${
      /** @type {Error***REMOVED*** */
      error.message***REMOVED***\n\n(@${name***REMOVED***)`);
    ***REMOVED***
  ***REMOVED***

  if (sourceMap) {
    postcssOptions.map = {
      annotation: false
    ***REMOVED***;
  ***REMOVED***
  /** @type {Postcss***REMOVED*** */
  // eslint-disable-next-line global-require


  const postcss = require("postcss").default; // @ts-ignore
  // eslint-disable-next-line global-require


  const cssnano = require("cssnano"); // @ts-ignore
  // Types are broken


  const result = await postcss([cssnano(minimizerOptions)]).process(code, postcssOptions);
  return {
    code: result.css,
    map: result.map ? result.map.toJSON() : // eslint-disable-next-line no-undefined
    undefined,
    warnings: result.warnings().map(String)
  ***REMOVED***;
***REMOVED***
/* istanbul ignore next */

/**
 * @param {Input***REMOVED*** input
 * @param {RawSourceMap | undefined***REMOVED*** sourceMap
 * @param {CustomOptions***REMOVED*** minimizerOptions
 * @return {Promise<MinimizedResult>***REMOVED***
 */


async function cssoMinify(input, sourceMap, minimizerOptions) {
  // eslint-disable-next-line global-require,import/no-extraneous-dependencies
  const csso = require("csso");

  const [[filename, code]] = Object.entries(input);
  const result = csso.minify(code, {
    filename,
    sourceMap: Boolean(sourceMap),
    ...minimizerOptions
  ***REMOVED***);
  return {
    code: result.css,
    map: result.map ?
    /** @type {SourceMapGenerator & { toJSON(): RawSourceMap ***REMOVED******REMOVED*** */
    result.map.toJSON() : // eslint-disable-next-line no-undefined
    undefined
  ***REMOVED***;
***REMOVED***
/* istanbul ignore next */

/**
 * @param {Input***REMOVED*** input
 * @param {RawSourceMap | undefined***REMOVED*** sourceMap
 * @param {CustomOptions***REMOVED*** minimizerOptions
 * @return {Promise<MinimizedResult>***REMOVED***
 */


async function cleanCssMinify(input, sourceMap, minimizerOptions) {
  // eslint-disable-next-line global-require,import/no-extraneous-dependencies
  const CleanCSS = require("clean-css");

  const [[name, code]] = Object.entries(input);
  const result = await new CleanCSS({
    sourceMap: Boolean(sourceMap),
    ...minimizerOptions,
    returnPromise: true
  ***REMOVED***).minify({
    [name]: {
      styles: code
    ***REMOVED***
  ***REMOVED***);
  const generatedSourceMap = result.sourceMap &&
  /** @type {SourceMapGenerator & { toJSON(): RawSourceMap ***REMOVED******REMOVED*** */
  result.sourceMap.toJSON(); // workaround for source maps on windows

  if (generatedSourceMap) {
    // eslint-disable-next-line global-require
    const isWindowsPathSep = require("path").sep === "\\";
    generatedSourceMap.sources = generatedSourceMap.sources.map(
    /**
     * @param {string***REMOVED*** item
     * @returns {string***REMOVED***
     */
    item => isWindowsPathSep ? item.replace(/\\/g, "/") : item);
  ***REMOVED***

  return {
    code: result.styles,
    map: generatedSourceMap,
    warnings: result.warnings
  ***REMOVED***;
***REMOVED***
/* istanbul ignore next */

/**
 * @param {Input***REMOVED*** input
 * @param {RawSourceMap | undefined***REMOVED*** sourceMap
 * @param {CustomOptions***REMOVED*** minimizerOptions
 * @return {Promise<MinimizedResult>***REMOVED***
 */


async function esbuildMinify(input, sourceMap, minimizerOptions) {
  /**
   * @param {import("esbuild").TransformOptions***REMOVED*** [esbuildOptions={***REMOVED***]
   * @returns {import("esbuild").TransformOptions***REMOVED***
   */
  const buildEsbuildOptions = (esbuildOptions = {***REMOVED***) => {
    // Need deep copy objects to avoid https://github.com/terser/terser/issues/366
    return {
      loader: "css",
      minify: true,
      legalComments: "inline",
      ...esbuildOptions,
      sourcemap: false
    ***REMOVED***;
  ***REMOVED***; // eslint-disable-next-line import/no-extraneous-dependencies, global-require


  const esbuild = require("esbuild"); // Copy `esbuild` options


  const esbuildOptions = buildEsbuildOptions(minimizerOptions); // Let `esbuild` generate a SourceMap

  if (sourceMap) {
    esbuildOptions.sourcemap = true;
    esbuildOptions.sourcesContent = false;
  ***REMOVED***

  const [[filename, code]] = Object.entries(input);
  esbuildOptions.sourcefile = filename;
  const result = await esbuild.transform(code, esbuildOptions);
  return {
    code: result.code,
    // eslint-disable-next-line no-undefined
    map: result.map ? JSON.parse(result.map) : undefined,
    warnings: result.warnings.length > 0 ? result.warnings.map(item => {
      return {
        source: item.location && item.location.file,
        // eslint-disable-next-line no-undefined
        line: item.location && item.location.line ? item.location.line : undefined,
        // eslint-disable-next-line no-undefined
        column: item.location && item.location.column ? item.location.column : undefined,
        plugin: item.pluginName,
        message: `${item.text***REMOVED***${item.detail ? `\nDetails:\n${item.detail***REMOVED***` : ""***REMOVED***${item.notes.length > 0 ? `\n\nNotes:\n${item.notes.map(note => `${note.location ? `[${note.location.file***REMOVED***:${note.location.line***REMOVED***:${note.location.column***REMOVED***] ` : ""***REMOVED***${note.text***REMOVED***${note.location ? `\nSuggestion: ${note.location.suggestion***REMOVED***` : ""***REMOVED***${note.location ? `\nLine text:\n${note.location.lineText***REMOVED***\n` : ""***REMOVED***`).join("\n")***REMOVED***` : ""***REMOVED***`
      ***REMOVED***;
    ***REMOVED***) : []
  ***REMOVED***;
***REMOVED***
/* istanbul ignore next */

/**
 * @param {Input***REMOVED*** input
 * @param {RawSourceMap | undefined***REMOVED*** sourceMap
 * @param {CustomOptions***REMOVED*** minimizerOptions
 * @return {Promise<MinimizedResult>***REMOVED***
 */


async function parcelCssMinify(input, sourceMap, minimizerOptions) {
  const [[filename, code]] = Object.entries(input);
  /**
   * @param {Partial<import("@parcel/css").TransformOptions>***REMOVED*** [parcelCssOptions={***REMOVED***]
   * @returns {import("@parcel/css").TransformOptions***REMOVED***
   */

  const buildParcelCssOptions = (parcelCssOptions = {***REMOVED***) => {
    // Need deep copy objects to avoid https://github.com/terser/terser/issues/366
    return {
      minify: true,
      ...parcelCssOptions,
      sourceMap: false,
      filename,
      code: Buffer.from(code)
    ***REMOVED***;
  ***REMOVED***; // eslint-disable-next-line import/no-extraneous-dependencies, global-require


  const parcelCss = require("@parcel/css"); // Copy `esbuild` options


  const parcelCssOptions = buildParcelCssOptions(minimizerOptions); // Let `esbuild` generate a SourceMap

  if (sourceMap) {
    parcelCssOptions.sourceMap = true;
  ***REMOVED***

  const result = await parcelCss.transform(parcelCssOptions);
  return {
    code: result.code.toString(),
    // eslint-disable-next-line no-undefined
    map: result.map ? JSON.parse(result.map.toString()) : undefined
  ***REMOVED***;
***REMOVED***

module.exports = {
  throttleAll,
  cssnanoMinify,
  cssoMinify,
  cleanCssMinify,
  esbuildMinify,
  parcelCssMinify
***REMOVED***;
