// @babel/eslint-parser/lib/worker/ast-info.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.getTokLabels = getTokLabels;
exports.getVisitorKeys = getVisitorKeys;
const _ESLINT_VISITOR_KEYS = require("eslint-visitor-keys");
const babel = require("./babel-core.cjs");
const ESLINT_VISITOR_KEYS = _ESLINT_VISITOR_KEYS.KEYS;
let visitorKeys;
function getVisitorKeys() {
  if (!visitorKeys) {
    const newTypes = {
      ChainExpression: ESLINT_VISITOR_KEYS.ChainExpression,
      ImportExpression: ESLINT_VISITOR_KEYS.ImportExpression,
      Literal: ESLINT_VISITOR_KEYS.Literal,
      MethodDefinition: ["decorators"].concat(ESLINT_VISITOR_KEYS.MethodDefinition),
      Property: ["decorators"].concat(ESLINT_VISITOR_KEYS.Property),
      PropertyDefinition: ["decorators", "typeAnnotation"].concat(ESLINT_VISITOR_KEYS.PropertyDefinition)
    ***REMOVED***;
    const conflictTypes = {
      ExportAllDeclaration: ESLINT_VISITOR_KEYS.ExportAllDeclaration
    ***REMOVED***;
    visitorKeys = Object.assign({***REMOVED***, newTypes, babel.types.VISITOR_KEYS, conflictTypes, {
      ClassPrivateMethod: ["decorators"].concat(ESLINT_VISITOR_KEYS.MethodDefinition)
    ***REMOVED***);
  ***REMOVED***
  return visitorKeys;
***REMOVED***
let tokLabels;
function getTokLabels() {
  return tokLabels || (tokLabels = (p => p.reduce((o, [k, v]) => Object.assign({***REMOVED***, o, {
    [k]: v
  ***REMOVED***), {***REMOVED***))((Object.entries || (o => Object.keys(o).map(k => [k, o[k]])))(babel.tokTypes).map(([key, tok]) => [key, tok.label])));
***REMOVED***
