// ajv/lib/vocabularies/applicator/additionalItems.ts
import type {
  CodeKeywordDefinition,
  ErrorObject,
  KeywordErrorDefinition,
  AnySchema,
***REMOVED*** from "../../types"
import type {KeywordCxt***REMOVED*** from "../../compile/validate"
import {_, str, not, Name***REMOVED*** from "../../compile/codegen"
import {alwaysValidSchema, checkStrictMode, Type***REMOVED*** from "../../compile/util"

export type AdditionalItemsError = ErrorObject<"additionalItems", {limit: number***REMOVED***, AnySchema>

const error: KeywordErrorDefinition = {
  message: ({params: {len***REMOVED******REMOVED***) => str`must NOT have more than ${len***REMOVED*** items`,
  params: ({params: {len***REMOVED******REMOVED***) => _`{limit: ${len***REMOVED******REMOVED***`,
***REMOVED***

const def: CodeKeywordDefinition = {
  keyword: "additionalItems" as const,
  type: "array",
  schemaType: ["boolean", "object"],
  before: "uniqueItems",
  error,
  code(cxt: KeywordCxt) {
    const {parentSchema, it***REMOVED*** = cxt
    const {items***REMOVED*** = parentSchema
    if (!Array.isArray(items)) {
      checkStrictMode(it, '"additionalItems" is ignored when "items" is not an array of schemas')
      return
    ***REMOVED***
    validateAdditionalItems(cxt, items)
  ***REMOVED***,
***REMOVED***

export function validateAdditionalItems(cxt: KeywordCxt, items: AnySchema[]): void {
  const {gen, schema, data, keyword, it***REMOVED*** = cxt
  it.items = true
  const len = gen.const("len", _`${data***REMOVED***.length`)
  if (schema === false) {
    cxt.setParams({len: items.length***REMOVED***)
    cxt.pass(_`${len***REMOVED*** <= ${items.length***REMOVED***`)
  ***REMOVED*** else if (typeof schema == "object" && !alwaysValidSchema(it, schema)) {
    const valid = gen.var("valid", _`${len***REMOVED*** <= ${items.length***REMOVED***`) // TODO var
    gen.if(not(valid), () => validateItems(valid))
    cxt.ok(valid)
  ***REMOVED***

  function validateItems(valid: Name): void {
    gen.forRange("i", items.length, len, (i) => {
      cxt.subschema({keyword, dataProp: i, dataPropType: Type.Num***REMOVED***, valid)
      if (!it.allErrors) gen.if(not(valid), () => gen.break())
    ***REMOVED***)
  ***REMOVED***
***REMOVED***

export default def
