// ajv/lib/vocabularies/applicator/items.ts
import type {CodeKeywordDefinition, AnySchema, AnySchemaObject***REMOVED*** from "../../types"
import type {KeywordCxt***REMOVED*** from "../../compile/validate"
import {_***REMOVED*** from "../../compile/codegen"
import {alwaysValidSchema, mergeEvaluated, checkStrictMode***REMOVED*** from "../../compile/util"
import {validateArray***REMOVED*** from "../code"

const def: CodeKeywordDefinition = {
  keyword: "items",
  type: "array",
  schemaType: ["object", "array", "boolean"],
  before: "uniqueItems",
  code(cxt: KeywordCxt) {
    const {schema, it***REMOVED*** = cxt
    if (Array.isArray(schema)) return validateTuple(cxt, "additionalItems", schema)
    it.items = true
    if (alwaysValidSchema(it, schema)) return
    cxt.ok(validateArray(cxt))
  ***REMOVED***,
***REMOVED***

export function validateTuple(
  cxt: KeywordCxt,
  extraItems: string,
  schArr: AnySchema[] = cxt.schema
): void {
  const {gen, parentSchema, data, keyword, it***REMOVED*** = cxt
  checkStrictTuple(parentSchema)
  if (it.opts.unevaluated && schArr.length && it.items !== true) {
    it.items = mergeEvaluated.items(gen, schArr.length, it.items)
  ***REMOVED***
  const valid = gen.name("valid")
  const len = gen.const("len", _`${data***REMOVED***.length`)
  schArr.forEach((sch: AnySchema, i: number) => {
    if (alwaysValidSchema(it, sch)) return
    gen.if(_`${len***REMOVED*** > ${i***REMOVED***`, () =>
      cxt.subschema(
        {
          keyword,
          schemaProp: i,
          dataProp: i,
        ***REMOVED***,
        valid
      )
    )
    cxt.ok(valid)
  ***REMOVED***)

  function checkStrictTuple(sch: AnySchemaObject): void {
    const {opts, errSchemaPath***REMOVED*** = it
    const l = schArr.length
    const fullTuple = l === sch.minItems && (l === sch.maxItems || sch[extraItems] === false)
    if (opts.strictTuples && !fullTuple) {
      const msg = `"${keyword***REMOVED***" is ${l***REMOVED***-tuple, but minItems or maxItems/${extraItems***REMOVED*** are not specified or different at path "${errSchemaPath***REMOVED***"`
      checkStrictMode(it, msg, opts.strictTuples)
    ***REMOVED***
  ***REMOVED***
***REMOVED***

export default def
