// @alloc/quick-lru/index.js
'use strict';

class QuickLRU {
	constructor(options = {***REMOVED***) {
		if (!(options.maxSize && options.maxSize > 0)) {
			throw new TypeError('`maxSize` must be a number greater than 0');
		***REMOVED***

		if (typeof options.maxAge === 'number' && options.maxAge === 0) {
			throw new TypeError('`maxAge` must be a number greater than 0');
		***REMOVED***

		this.maxSize = options.maxSize;
		this.maxAge = options.maxAge || Infinity;
		this.onEviction = options.onEviction;
		this.cache = new Map();
		this.oldCache = new Map();
		this._size = 0;
	***REMOVED***

	_emitEvictions(cache) {
		if (typeof this.onEviction !== 'function') {
			return;
		***REMOVED***

		for (const [key, item] of cache) {
			this.onEviction(key, item.value);
		***REMOVED***
	***REMOVED***

	_deleteIfExpired(key, item) {
		if (typeof item.expiry === 'number' && item.expiry <= Date.now()) {
			if (typeof this.onEviction === 'function') {
				this.onEviction(key, item.value);
			***REMOVED***

			return this.delete(key);
		***REMOVED***

		return false;
	***REMOVED***

	_getOrDeleteIfExpired(key, item) {
		const deleted = this._deleteIfExpired(key, item);
		if (deleted === false) {
			return item.value;
		***REMOVED***
	***REMOVED***

	_getItemValue(key, item) {
		return item.expiry ? this._getOrDeleteIfExpired(key, item) : item.value;
	***REMOVED***

	_peek(key, cache) {
		const item = cache.get(key);

		return this._getItemValue(key, item);
	***REMOVED***

	_set(key, value) {
		this.cache.set(key, value);
		this._size++;

		if (this._size >= this.maxSize) {
			this._size = 0;
			this._emitEvictions(this.oldCache);
			this.oldCache = this.cache;
			this.cache = new Map();
		***REMOVED***
	***REMOVED***

	_moveToRecent(key, item) {
		this.oldCache.delete(key);
		this._set(key, item);
	***REMOVED***

	* _entriesAscending() {
		for (const item of this.oldCache) {
			const [key, value] = item;
			if (!this.cache.has(key)) {
				const deleted = this._deleteIfExpired(key, value);
				if (deleted === false) {
					yield item;
				***REMOVED***
			***REMOVED***
		***REMOVED***

		for (const item of this.cache) {
			const [key, value] = item;
			const deleted = this._deleteIfExpired(key, value);
			if (deleted === false) {
				yield item;
			***REMOVED***
		***REMOVED***
	***REMOVED***

	get(key) {
		if (this.cache.has(key)) {
			const item = this.cache.get(key);

			return this._getItemValue(key, item);
		***REMOVED***

		if (this.oldCache.has(key)) {
			const item = this.oldCache.get(key);
			if (this._deleteIfExpired(key, item) === false) {
				this._moveToRecent(key, item);
				return item.value;
			***REMOVED***
		***REMOVED***
	***REMOVED***

	set(key, value, {maxAge = this.maxAge === Infinity ? undefined : Date.now() + this.maxAge***REMOVED*** = {***REMOVED***) {
		if (this.cache.has(key)) {
			this.cache.set(key, {
				value,
				maxAge
			***REMOVED***);
		***REMOVED*** else {
			this._set(key, {value, expiry: maxAge***REMOVED***);
		***REMOVED***
	***REMOVED***

	has(key) {
		if (this.cache.has(key)) {
			return !this._deleteIfExpired(key, this.cache.get(key));
		***REMOVED***

		if (this.oldCache.has(key)) {
			return !this._deleteIfExpired(key, this.oldCache.get(key));
		***REMOVED***

		return false;
	***REMOVED***

	peek(key) {
		if (this.cache.has(key)) {
			return this._peek(key, this.cache);
		***REMOVED***

		if (this.oldCache.has(key)) {
			return this._peek(key, this.oldCache);
		***REMOVED***
	***REMOVED***

	delete(key) {
		const deleted = this.cache.delete(key);
		if (deleted) {
			this._size--;
		***REMOVED***

		return this.oldCache.delete(key) || deleted;
	***REMOVED***

	clear() {
		this.cache.clear();
		this.oldCache.clear();
		this._size = 0;
	***REMOVED***
	
	resize(newSize) {
		if (!(newSize && newSize > 0)) {
			throw new TypeError('`maxSize` must be a number greater than 0');
		***REMOVED***

		const items = [...this._entriesAscending()];
		const removeCount = items.length - newSize;
		if (removeCount < 0) {
			this.cache = new Map(items);
			this.oldCache = new Map();
			this._size = items.length;
		***REMOVED*** else {
			if (removeCount > 0) {
				this._emitEvictions(items.slice(0, removeCount));
			***REMOVED***

			this.oldCache = new Map(items.slice(removeCount));
			this.cache = new Map();
			this._size = 0;
		***REMOVED***

		this.maxSize = newSize;
	***REMOVED***

	* keys() {
		for (const [key] of this) {
			yield key;
		***REMOVED***
	***REMOVED***

	* values() {
		for (const [, value] of this) {
			yield value;
		***REMOVED***
	***REMOVED***

	* [Symbol.iterator]() {
		for (const item of this.cache) {
			const [key, value] = item;
			const deleted = this._deleteIfExpired(key, value);
			if (deleted === false) {
				yield [key, value.value];
			***REMOVED***
		***REMOVED***

		for (const item of this.oldCache) {
			const [key, value] = item;
			if (!this.cache.has(key)) {
				const deleted = this._deleteIfExpired(key, value);
				if (deleted === false) {
					yield [key, value.value];
				***REMOVED***
			***REMOVED***
		***REMOVED***
	***REMOVED***

	* entriesDescending() {
		let items = [...this.cache];
		for (let i = items.length - 1; i >= 0; --i) {
			const item = items[i];
			const [key, value] = item;
			const deleted = this._deleteIfExpired(key, value);
			if (deleted === false) {
				yield [key, value.value];
			***REMOVED***
		***REMOVED***

		items = [...this.oldCache];
		for (let i = items.length - 1; i >= 0; --i) {
			const item = items[i];
			const [key, value] = item;
			if (!this.cache.has(key)) {
				const deleted = this._deleteIfExpired(key, value);
				if (deleted === false) {
					yield [key, value.value];
				***REMOVED***
			***REMOVED***
		***REMOVED***
	***REMOVED***

	* entriesAscending() {
		for (const [key, value] of this._entriesAscending()) {
			yield [key, value.value];
		***REMOVED***
	***REMOVED***

	get size() {
		if (!this._size) {
			return this.oldCache.size;
		***REMOVED***

		let oldCacheSize = 0;
		for (const key of this.oldCache.keys()) {
			if (!this.cache.has(key)) {
				oldCacheSize++;
			***REMOVED***
		***REMOVED***

		return Math.min(this._size + oldCacheSize, this.maxSize);
	***REMOVED***
***REMOVED***

module.exports = QuickLRU;
