// domexception/lib/DOMException.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const impl = utils.implSymbol;
const ctorRegistry = utils.ctorRegistrySymbol;

const iface = {
  // When an interface-module that implements this interface as a mixin is loaded, it will append its own `.is()`
  // method into this array. It allows objects that directly implements *those* interfaces to be recognized as
  // implementing this mixin interface.
  _mixedIntoPredicates: [],
  is(obj) {
    if (obj) {
      if (utils.hasOwn(obj, impl) && obj[impl] instanceof Impl.implementation) {
        return true;
      ***REMOVED***
      for (const isMixedInto of module.exports._mixedIntoPredicates) {
        if (isMixedInto(obj)) {
          return true;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
    return false;
  ***REMOVED***,
  isImpl(obj) {
    if (obj) {
      if (obj instanceof Impl.implementation) {
        return true;
      ***REMOVED***

      const wrapper = utils.wrapperForImpl(obj);
      for (const isMixedInto of module.exports._mixedIntoPredicates) {
        if (isMixedInto(wrapper)) {
          return true;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
    return false;
  ***REMOVED***,
  convert(obj, { context = "The provided value" ***REMOVED*** = {***REMOVED***) {
    if (module.exports.is(obj)) {
      return utils.implForWrapper(obj);
    ***REMOVED***
    throw new TypeError(`${context***REMOVED*** is not of type 'DOMException'.`);
  ***REMOVED***,

  create(globalObject, constructorArgs, privateData) {
    if (globalObject[ctorRegistry] === undefined) {
      throw new Error("Internal error: invalid global object");
    ***REMOVED***

    const ctor = globalObject[ctorRegistry]["DOMException"];
    if (ctor === undefined) {
      throw new Error("Internal error: constructor DOMException is not installed on the passed global object");
    ***REMOVED***

    let obj = Object.create(ctor.prototype);
    obj = iface.setup(obj, globalObject, constructorArgs, privateData);
    return obj;
  ***REMOVED***,
  createImpl(globalObject, constructorArgs, privateData) {
    const obj = iface.create(globalObject, constructorArgs, privateData);
    return utils.implForWrapper(obj);
  ***REMOVED***,
  _internalSetup(obj) {***REMOVED***,
  setup(obj, globalObject, constructorArgs = [], privateData = {***REMOVED***) {
    privateData.wrapper = obj;

    iface._internalSetup(obj);
    Object.defineProperty(obj, impl, {
      value: new Impl.implementation(globalObject, constructorArgs, privateData),
      configurable: true
    ***REMOVED***);

    obj[impl][utils.wrapperSymbol] = obj;
    if (Impl.init) {
      Impl.init(obj[impl], privateData);
    ***REMOVED***
    return obj;
  ***REMOVED***,

  install(globalObject) {
    class DOMException {
      constructor() {
        const args = [];
        {
          let curArg = arguments[0];
          if (curArg !== undefined) {
            curArg = conversions["DOMString"](curArg, { context: "Failed to construct 'DOMException': parameter 1" ***REMOVED***);
          ***REMOVED*** else {
            curArg = "";
          ***REMOVED***
          args.push(curArg);
        ***REMOVED***
        {
          let curArg = arguments[1];
          if (curArg !== undefined) {
            curArg = conversions["DOMString"](curArg, { context: "Failed to construct 'DOMException': parameter 2" ***REMOVED***);
          ***REMOVED*** else {
            curArg = "Error";
          ***REMOVED***
          args.push(curArg);
        ***REMOVED***
        return iface.setup(Object.create(new.target.prototype), globalObject, args);
      ***REMOVED***

      get name() {
        if (!this || !module.exports.is(this)) {
          throw new TypeError("Illegal invocation");
        ***REMOVED***

        return this[impl]["name"];
      ***REMOVED***

      get message() {
        if (!this || !module.exports.is(this)) {
          throw new TypeError("Illegal invocation");
        ***REMOVED***

        return this[impl]["message"];
      ***REMOVED***

      get code() {
        if (!this || !module.exports.is(this)) {
          throw new TypeError("Illegal invocation");
        ***REMOVED***

        return this[impl]["code"];
      ***REMOVED***
    ***REMOVED***
    Object.defineProperties(DOMException.prototype, {
      name: { enumerable: true ***REMOVED***,
      message: { enumerable: true ***REMOVED***,
      code: { enumerable: true ***REMOVED***,
      [Symbol.toStringTag]: { value: "DOMException", configurable: true ***REMOVED***,
      INDEX_SIZE_ERR: { value: 1, enumerable: true ***REMOVED***,
      DOMSTRING_SIZE_ERR: { value: 2, enumerable: true ***REMOVED***,
      HIERARCHY_REQUEST_ERR: { value: 3, enumerable: true ***REMOVED***,
      WRONG_DOCUMENT_ERR: { value: 4, enumerable: true ***REMOVED***,
      INVALID_CHARACTER_ERR: { value: 5, enumerable: true ***REMOVED***,
      NO_DATA_ALLOWED_ERR: { value: 6, enumerable: true ***REMOVED***,
      NO_MODIFICATION_ALLOWED_ERR: { value: 7, enumerable: true ***REMOVED***,
      NOT_FOUND_ERR: { value: 8, enumerable: true ***REMOVED***,
      NOT_SUPPORTED_ERR: { value: 9, enumerable: true ***REMOVED***,
      INUSE_ATTRIBUTE_ERR: { value: 10, enumerable: true ***REMOVED***,
      INVALID_STATE_ERR: { value: 11, enumerable: true ***REMOVED***,
      SYNTAX_ERR: { value: 12, enumerable: true ***REMOVED***,
      INVALID_MODIFICATION_ERR: { value: 13, enumerable: true ***REMOVED***,
      NAMESPACE_ERR: { value: 14, enumerable: true ***REMOVED***,
      INVALID_ACCESS_ERR: { value: 15, enumerable: true ***REMOVED***,
      VALIDATION_ERR: { value: 16, enumerable: true ***REMOVED***,
      TYPE_MISMATCH_ERR: { value: 17, enumerable: true ***REMOVED***,
      SECURITY_ERR: { value: 18, enumerable: true ***REMOVED***,
      NETWORK_ERR: { value: 19, enumerable: true ***REMOVED***,
      ABORT_ERR: { value: 20, enumerable: true ***REMOVED***,
      URL_MISMATCH_ERR: { value: 21, enumerable: true ***REMOVED***,
      QUOTA_EXCEEDED_ERR: { value: 22, enumerable: true ***REMOVED***,
      TIMEOUT_ERR: { value: 23, enumerable: true ***REMOVED***,
      INVALID_NODE_TYPE_ERR: { value: 24, enumerable: true ***REMOVED***,
      DATA_CLONE_ERR: { value: 25, enumerable: true ***REMOVED***
    ***REMOVED***);
    Object.defineProperties(DOMException, {
      INDEX_SIZE_ERR: { value: 1, enumerable: true ***REMOVED***,
      DOMSTRING_SIZE_ERR: { value: 2, enumerable: true ***REMOVED***,
      HIERARCHY_REQUEST_ERR: { value: 3, enumerable: true ***REMOVED***,
      WRONG_DOCUMENT_ERR: { value: 4, enumerable: true ***REMOVED***,
      INVALID_CHARACTER_ERR: { value: 5, enumerable: true ***REMOVED***,
      NO_DATA_ALLOWED_ERR: { value: 6, enumerable: true ***REMOVED***,
      NO_MODIFICATION_ALLOWED_ERR: { value: 7, enumerable: true ***REMOVED***,
      NOT_FOUND_ERR: { value: 8, enumerable: true ***REMOVED***,
      NOT_SUPPORTED_ERR: { value: 9, enumerable: true ***REMOVED***,
      INUSE_ATTRIBUTE_ERR: { value: 10, enumerable: true ***REMOVED***,
      INVALID_STATE_ERR: { value: 11, enumerable: true ***REMOVED***,
      SYNTAX_ERR: { value: 12, enumerable: true ***REMOVED***,
      INVALID_MODIFICATION_ERR: { value: 13, enumerable: true ***REMOVED***,
      NAMESPACE_ERR: { value: 14, enumerable: true ***REMOVED***,
      INVALID_ACCESS_ERR: { value: 15, enumerable: true ***REMOVED***,
      VALIDATION_ERR: { value: 16, enumerable: true ***REMOVED***,
      TYPE_MISMATCH_ERR: { value: 17, enumerable: true ***REMOVED***,
      SECURITY_ERR: { value: 18, enumerable: true ***REMOVED***,
      NETWORK_ERR: { value: 19, enumerable: true ***REMOVED***,
      ABORT_ERR: { value: 20, enumerable: true ***REMOVED***,
      URL_MISMATCH_ERR: { value: 21, enumerable: true ***REMOVED***,
      QUOTA_EXCEEDED_ERR: { value: 22, enumerable: true ***REMOVED***,
      TIMEOUT_ERR: { value: 23, enumerable: true ***REMOVED***,
      INVALID_NODE_TYPE_ERR: { value: 24, enumerable: true ***REMOVED***,
      DATA_CLONE_ERR: { value: 25, enumerable: true ***REMOVED***
    ***REMOVED***);
    if (globalObject[ctorRegistry] === undefined) {
      globalObject[ctorRegistry] = Object.create(null);
    ***REMOVED***
    globalObject[ctorRegistry]["DOMException"] = DOMException;

    Object.defineProperty(globalObject, "DOMException", {
      configurable: true,
      writable: true,
      value: DOMException
    ***REMOVED***);
  ***REMOVED***
***REMOVED***; // iface
module.exports = iface;

const Impl = require("./DOMException-impl.js");
