// csso/lib/compress.js
var List = require('css-tree').List;
var clone = require('css-tree').clone;
var usageUtils = require('./usage');
var clean = require('./clean');
var replace = require('./replace');
var restructure = require('./restructure');
var walk = require('css-tree').walk;

function readChunk(children, specialComments) {
    var buffer = new List();
    var nonSpaceTokenInBuffer = false;
    var protectedComment;

    children.nextUntil(children.head, function(node, item, list) {
        if (node.type === 'Comment') {
            if (!specialComments || node.value.charAt(0) !== '!') {
                list.remove(item);
                return;
            ***REMOVED***

            if (nonSpaceTokenInBuffer || protectedComment) {
                return true;
            ***REMOVED***

            list.remove(item);
            protectedComment = node;
            return;
        ***REMOVED***

        if (node.type !== 'WhiteSpace') {
            nonSpaceTokenInBuffer = true;
        ***REMOVED***

        buffer.insert(list.remove(item));
    ***REMOVED***);

    return {
        comment: protectedComment,
        stylesheet: {
            type: 'StyleSheet',
            loc: null,
            children: buffer
        ***REMOVED***
    ***REMOVED***;
***REMOVED***

function compressChunk(ast, firstAtrulesAllowed, num, options) {
    options.logger('Compress block #' + num, null, true);

    var seed = 1;

    if (ast.type === 'StyleSheet') {
        ast.firstAtrulesAllowed = firstAtrulesAllowed;
        ast.id = seed++;
    ***REMOVED***

    walk(ast, {
        visit: 'Atrule',
        enter: function markScopes(node) {
            if (node.block !== null) {
                node.block.id = seed++;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***);
    options.logger('init', ast);

    // remove redundant
    clean(ast, options);
    options.logger('clean', ast);

    // replace nodes for shortened forms
    replace(ast, options);
    options.logger('replace', ast);

    // structure optimisations
    if (options.restructuring) {
        restructure(ast, options);
    ***REMOVED***

    return ast;
***REMOVED***

function getCommentsOption(options) {
    var comments = 'comments' in options ? options.comments : 'exclamation';

    if (typeof comments === 'boolean') {
        comments = comments ? 'exclamation' : false;
    ***REMOVED*** else if (comments !== 'exclamation' && comments !== 'first-exclamation') {
        comments = false;
    ***REMOVED***

    return comments;
***REMOVED***

function getRestructureOption(options) {
    if ('restructure' in options) {
        return options.restructure;
    ***REMOVED***

    return 'restructuring' in options ? options.restructuring : true;
***REMOVED***

function wrapBlock(block) {
    return new List().appendData({
        type: 'Rule',
        loc: null,
        prelude: {
            type: 'SelectorList',
            loc: null,
            children: new List().appendData({
                type: 'Selector',
                loc: null,
                children: new List().appendData({
                    type: 'TypeSelector',
                    loc: null,
                    name: 'x'
                ***REMOVED***)
            ***REMOVED***)
        ***REMOVED***,
        block: block
    ***REMOVED***);
***REMOVED***

module.exports = function compress(ast, options) {
    ast = ast || { type: 'StyleSheet', loc: null, children: new List() ***REMOVED***;
    options = options || {***REMOVED***;

    var compressOptions = {
        logger: typeof options.logger === 'function' ? options.logger : function() {***REMOVED***,
        restructuring: getRestructureOption(options),
        forceMediaMerge: Boolean(options.forceMediaMerge),
        usage: options.usage ? usageUtils.buildIndex(options.usage) : false
    ***REMOVED***;
    var specialComments = getCommentsOption(options);
    var firstAtrulesAllowed = true;
    var input;
    var output = new List();
    var chunk;
    var chunkNum = 1;
    var chunkChildren;

    if (options.clone) {
        ast = clone(ast);
    ***REMOVED***

    if (ast.type === 'StyleSheet') {
        input = ast.children;
        ast.children = output;
    ***REMOVED*** else {
        input = wrapBlock(ast);
    ***REMOVED***

    do {
        chunk = readChunk(input, Boolean(specialComments));
        compressChunk(chunk.stylesheet, firstAtrulesAllowed, chunkNum++, compressOptions);
        chunkChildren = chunk.stylesheet.children;

        if (chunk.comment) {
            // add \n before comment if there is another content in output
            if (!output.isEmpty()) {
                output.insert(List.createItem({
                    type: 'Raw',
                    value: '\n'
                ***REMOVED***));
            ***REMOVED***

            output.insert(List.createItem(chunk.comment));

            // add \n after comment if chunk is not empty
            if (!chunkChildren.isEmpty()) {
                output.insert(List.createItem({
                    type: 'Raw',
                    value: '\n'
                ***REMOVED***));
            ***REMOVED***
        ***REMOVED***

        if (firstAtrulesAllowed && !chunkChildren.isEmpty()) {
            var lastRule = chunkChildren.last();

            if (lastRule.type !== 'Atrule' ||
               (lastRule.name !== 'import' && lastRule.name !== 'charset')) {
                firstAtrulesAllowed = false;
            ***REMOVED***
        ***REMOVED***

        if (specialComments !== 'exclamation') {
            specialComments = false;
        ***REMOVED***

        output.appendList(chunkChildren);
    ***REMOVED*** while (!input.isEmpty());

    return {
        ast: ast
    ***REMOVED***;
***REMOVED***;
