// csso/lib/index.js
var csstree = require('css-tree');
var parse = csstree.parse;
var compress = require('./compress');
var generate = csstree.generate;

function debugOutput(name, options, startTime, data) {
    if (options.debug) {
        console.error('## ' + name + ' done in %d ms\n', Date.now() - startTime);
    ***REMOVED***

    return data;
***REMOVED***

function createDefaultLogger(level) {
    var lastDebug;

    return function logger(title, ast) {
        var line = title;

        if (ast) {
            line = '[' + ((Date.now() - lastDebug) / 1000).toFixed(3) + 's] ' + line;
        ***REMOVED***

        if (level > 1 && ast) {
            var css = generate(ast);

            // when level 2, limit css to 256 symbols
            if (level === 2 && css.length > 256) {
                css = css.substr(0, 256) + '...';
            ***REMOVED***

            line += '\n  ' + css + '\n';
        ***REMOVED***

        console.error(line);
        lastDebug = Date.now();
    ***REMOVED***;
***REMOVED***

function copy(obj) {
    var result = {***REMOVED***;

    for (var key in obj) {
        result[key] = obj[key];
    ***REMOVED***

    return result;
***REMOVED***

function buildCompressOptions(options) {
    options = copy(options);

    if (typeof options.logger !== 'function' && options.debug) {
        options.logger = createDefaultLogger(options.debug);
    ***REMOVED***

    return options;
***REMOVED***

function runHandler(ast, options, handlers) {
    if (!Array.isArray(handlers)) {
        handlers = [handlers];
    ***REMOVED***

    handlers.forEach(function(fn) {
        fn(ast, options);
    ***REMOVED***);
***REMOVED***

function minify(context, source, options) {
    options = options || {***REMOVED***;

    var filename = options.filename || '<unknown>';
    var result;

    // parse
    var ast = debugOutput('parsing', options, Date.now(),
        parse(source, {
            context: context,
            filename: filename,
            positions: Boolean(options.sourceMap)
        ***REMOVED***)
    );

    // before compress handlers
    if (options.beforeCompress) {
        debugOutput('beforeCompress', options, Date.now(),
            runHandler(ast, options, options.beforeCompress)
        );
    ***REMOVED***

    // compress
    var compressResult = debugOutput('compress', options, Date.now(),
        compress(ast, buildCompressOptions(options))
    );

    // after compress handlers
    if (options.afterCompress) {
        debugOutput('afterCompress', options, Date.now(),
            runHandler(compressResult, options, options.afterCompress)
        );
    ***REMOVED***

    // generate
    if (options.sourceMap) {
        result = debugOutput('generate(sourceMap: true)', options, Date.now(), (function() {
            var tmp = generate(compressResult.ast, { sourceMap: true ***REMOVED***);
            tmp.map._file = filename; // since other tools can relay on file in source map transform chain
            tmp.map.setSourceContent(filename, source);
            return tmp;
        ***REMOVED***()));
    ***REMOVED*** else {
        result = debugOutput('generate', options, Date.now(), {
            css: generate(compressResult.ast),
            map: null
        ***REMOVED***);
    ***REMOVED***

    return result;
***REMOVED***

function minifyStylesheet(source, options) {
    return minify('stylesheet', source, options);
***REMOVED***

function minifyBlock(source, options) {
    return minify('declarationList', source, options);
***REMOVED***

module.exports = {
    version: require('../package.json').version,

    // main methods
    minify: minifyStylesheet,
    minifyBlock: minifyBlock,

    // css syntax parser/walkers/generator/etc
    syntax: Object.assign({
        compress: compress
    ***REMOVED***, csstree)
***REMOVED***;
