// csso/lib/usage.js
var hasOwnProperty = Object.prototype.hasOwnProperty;

function buildMap(list, caseInsensitive) {
    var map = Object.create(null);

    if (!Array.isArray(list)) {
        return null;
    ***REMOVED***

    for (var i = 0; i < list.length; i++) {
        var name = list[i];

        if (caseInsensitive) {
            name = name.toLowerCase();
        ***REMOVED***

        map[name] = true;
    ***REMOVED***

    return map;
***REMOVED***

function buildList(data) {
    if (!data) {
        return null;
    ***REMOVED***

    var tags = buildMap(data.tags, true);
    var ids = buildMap(data.ids);
    var classes = buildMap(data.classes);

    if (tags === null &&
        ids === null &&
        classes === null) {
        return null;
    ***REMOVED***

    return {
        tags: tags,
        ids: ids,
        classes: classes
    ***REMOVED***;
***REMOVED***

function buildIndex(data) {
    var scopes = false;

    if (data.scopes && Array.isArray(data.scopes)) {
        scopes = Object.create(null);

        for (var i = 0; i < data.scopes.length; i++) {
            var list = data.scopes[i];

            if (!list || !Array.isArray(list)) {
                throw new Error('Wrong usage format');
            ***REMOVED***

            for (var j = 0; j < list.length; j++) {
                var name = list[j];

                if (hasOwnProperty.call(scopes, name)) {
                    throw new Error('Class can\'t be used for several scopes: ' + name);
                ***REMOVED***

                scopes[name] = i + 1;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    return {
        whitelist: buildList(data),
        blacklist: buildList(data.blacklist),
        scopes: scopes
    ***REMOVED***;
***REMOVED***

module.exports = {
    buildIndex: buildIndex
***REMOVED***;
