// @rollup/plugin-node-resolve/dist/cjs/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true ***REMOVED***);

var path = require('path');
var builtinList = require('builtin-modules');
var deepMerge = require('deepmerge');
var isModule = require('is-module');
var fs = require('fs');
var util = require('util');
var url = require('url');
var resolve = require('resolve');
var pluginutils = require('@rollup/pluginutils');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e ***REMOVED***; ***REMOVED***

var path__default = /*#__PURE__*/_interopDefaultLegacy(path);
var builtinList__default = /*#__PURE__*/_interopDefaultLegacy(builtinList);
var deepMerge__default = /*#__PURE__*/_interopDefaultLegacy(deepMerge);
var isModule__default = /*#__PURE__*/_interopDefaultLegacy(isModule);
var fs__default = /*#__PURE__*/_interopDefaultLegacy(fs);
var resolve__default = /*#__PURE__*/_interopDefaultLegacy(resolve);

const access = util.promisify(fs__default['default'].access);
const readFile = util.promisify(fs__default['default'].readFile);
const realpath = util.promisify(fs__default['default'].realpath);
const stat = util.promisify(fs__default['default'].stat);
async function exists(filePath) {
  try {
    await access(filePath);
    return true;
  ***REMOVED*** catch {
    return false;
  ***REMOVED***
***REMOVED***

const onError = (error) => {
  if (error.code === 'ENOENT') {
    return false;
  ***REMOVED***
  throw error;
***REMOVED***;

const makeCache = (fn) => {
  const cache = new Map();
  const wrapped = async (param, done) => {
    if (cache.has(param) === false) {
      cache.set(
        param,
        fn(param).catch((err) => {
          cache.delete(param);
          throw err;
        ***REMOVED***)
      );
    ***REMOVED***

    try {
      const result = cache.get(param);
      const value = await result;
      return done(null, value);
    ***REMOVED*** catch (error) {
      return done(error);
    ***REMOVED***
  ***REMOVED***;

  wrapped.clear = () => cache.clear();

  return wrapped;
***REMOVED***;

const isDirCached = makeCache(async (file) => {
  try {
    const stats = await stat(file);
    return stats.isDirectory();
  ***REMOVED*** catch (error) {
    return onError(error);
  ***REMOVED***
***REMOVED***);

const isFileCached = makeCache(async (file) => {
  try {
    const stats = await stat(file);
    return stats.isFile();
  ***REMOVED*** catch (error) {
    return onError(error);
  ***REMOVED***
***REMOVED***);

const readCachedFile = makeCache(readFile);

// returns the imported package name for bare module imports
function getPackageName(id) {
  if (id.startsWith('.') || id.startsWith('/')) {
    return null;
  ***REMOVED***

  const split = id.split('/');

  // @my-scope/my-package/foo.js -> @my-scope/my-package
  // @my-scope/my-package -> @my-scope/my-package
  if (split[0][0] === '@') {
    return `${split[0]***REMOVED***/${split[1]***REMOVED***`;
  ***REMOVED***

  // my-package/foo.js -> my-package
  // my-package -> my-package
  return split[0];
***REMOVED***

function getMainFields(options) {
  let mainFields;
  if (options.mainFields) {
    ({ mainFields ***REMOVED*** = options);
  ***REMOVED*** else {
    mainFields = ['module', 'main'];
  ***REMOVED***
  if (options.browser && mainFields.indexOf('browser') === -1) {
    return ['browser'].concat(mainFields);
  ***REMOVED***
  if (!mainFields.length) {
    throw new Error('Please ensure at least one `mainFields` value is specified');
  ***REMOVED***
  return mainFields;
***REMOVED***

function getPackageInfo(options) {
  const {
    cache,
    extensions,
    pkg,
    mainFields,
    preserveSymlinks,
    useBrowserOverrides,
    rootDir,
    ignoreSideEffectsForRoot
  ***REMOVED*** = options;
  let { pkgPath ***REMOVED*** = options;

  if (cache.has(pkgPath)) {
    return cache.get(pkgPath);
  ***REMOVED***

  // browserify/resolve doesn't realpath paths returned in its packageFilter callback
  if (!preserveSymlinks) {
    pkgPath = fs.realpathSync(pkgPath);
  ***REMOVED***

  const pkgRoot = path.dirname(pkgPath);

  const packageInfo = {
    // copy as we are about to munge the `main` field of `pkg`.
    packageJson: { ...pkg ***REMOVED***,

    // path to package.json file
    packageJsonPath: pkgPath,

    // directory containing the package.json
    root: pkgRoot,

    // which main field was used during resolution of this module (main, module, or browser)
    resolvedMainField: 'main',

    // whether the browser map was used to resolve the entry point to this module
    browserMappedMain: false,

    // the entry point of the module with respect to the selected main field and any
    // relevant browser mappings.
    resolvedEntryPoint: ''
  ***REMOVED***;

  let overriddenMain = false;
  for (let i = 0; i < mainFields.length; i++) {
    const field = mainFields[i];
    if (typeof pkg[field] === 'string') {
      pkg.main = pkg[field];
      packageInfo.resolvedMainField = field;
      overriddenMain = true;
      break;
    ***REMOVED***
  ***REMOVED***

  const internalPackageInfo = {
    cachedPkg: pkg,
    hasModuleSideEffects: () => null,
    hasPackageEntry: overriddenMain !== false || mainFields.indexOf('main') !== -1,
    packageBrowserField:
      useBrowserOverrides &&
      typeof pkg.browser === 'object' &&
      Object.keys(pkg.browser).reduce((browser, key) => {
        let resolved = pkg.browser[key];
        if (resolved && resolved[0] === '.') {
          resolved = path.resolve(pkgRoot, resolved);
        ***REMOVED***
        /* eslint-disable no-param-reassign */
        browser[key] = resolved;
        if (key[0] === '.') {
          const absoluteKey = path.resolve(pkgRoot, key);
          browser[absoluteKey] = resolved;
          if (!path.extname(key)) {
            extensions.reduce((subBrowser, ext) => {
              subBrowser[absoluteKey + ext] = subBrowser[key];
              return subBrowser;
            ***REMOVED***, browser);
          ***REMOVED***
        ***REMOVED***
        return browser;
      ***REMOVED***, {***REMOVED***),
    packageInfo
  ***REMOVED***;

  const browserMap = internalPackageInfo.packageBrowserField;
  if (
    useBrowserOverrides &&
    typeof pkg.browser === 'object' &&
    // eslint-disable-next-line no-prototype-builtins
    browserMap.hasOwnProperty(pkg.main)
  ) {
    packageInfo.resolvedEntryPoint = browserMap[pkg.main];
    packageInfo.browserMappedMain = true;
  ***REMOVED*** else {
    // index.node is technically a valid default entrypoint as well...
    packageInfo.resolvedEntryPoint = path.resolve(pkgRoot, pkg.main || 'index.js');
    packageInfo.browserMappedMain = false;
  ***REMOVED***

  if (!ignoreSideEffectsForRoot || rootDir !== pkgRoot) {
    const packageSideEffects = pkg.sideEffects;
    if (typeof packageSideEffects === 'boolean') {
      internalPackageInfo.hasModuleSideEffects = () => packageSideEffects;
    ***REMOVED*** else if (Array.isArray(packageSideEffects)) {
      internalPackageInfo.hasModuleSideEffects = pluginutils.createFilter(packageSideEffects, null, {
        resolve: pkgRoot
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***

  cache.set(pkgPath, internalPackageInfo);
  return internalPackageInfo;
***REMOVED***

function normalizeInput(input) {
  if (Array.isArray(input)) {
    return input;
  ***REMOVED*** else if (typeof input === 'object') {
    return Object.values(input);
  ***REMOVED***

  // otherwise it's a string
  return [input];
***REMOVED***

/* eslint-disable no-await-in-loop */

const fileExists = util.promisify(fs__default['default'].exists);

function isModuleDir(current, moduleDirs) {
  return moduleDirs.some((dir) => current.endsWith(dir));
***REMOVED***

async function findPackageJson(base, moduleDirs) {
  const { root ***REMOVED*** = path__default['default'].parse(base);
  let current = base;

  while (current !== root && !isModuleDir(current, moduleDirs)) {
    const pkgJsonPath = path__default['default'].join(current, 'package.json');
    if (await fileExists(pkgJsonPath)) {
      const pkgJsonString = fs__default['default'].readFileSync(pkgJsonPath, 'utf-8');
      return { pkgJson: JSON.parse(pkgJsonString), pkgPath: current, pkgJsonPath ***REMOVED***;
    ***REMOVED***
    current = path__default['default'].resolve(current, '..');
  ***REMOVED***
  return null;
***REMOVED***

function isUrl(str) {
  try {
    return !!new URL(str);
  ***REMOVED*** catch (_) {
    return false;
  ***REMOVED***
***REMOVED***

function isConditions(exports) {
  return typeof exports === 'object' && Object.keys(exports).every((k) => !k.startsWith('.'));
***REMOVED***

function isMappings(exports) {
  return typeof exports === 'object' && !isConditions(exports);
***REMOVED***

function isMixedExports(exports) {
  const keys = Object.keys(exports);
  return keys.some((k) => k.startsWith('.')) && keys.some((k) => !k.startsWith('.'));
***REMOVED***

function createBaseErrorMsg(importSpecifier, importer) {
  return `Could not resolve import "${importSpecifier***REMOVED***" in ${importer***REMOVED***`;
***REMOVED***

function createErrorMsg(context, reason, internal) {
  const { importSpecifier, importer, pkgJsonPath ***REMOVED*** = context;
  const base = createBaseErrorMsg(importSpecifier, importer);
  const field = internal ? 'imports' : 'exports';
  return `${base***REMOVED*** using ${field***REMOVED*** defined in ${pkgJsonPath***REMOVED***.${reason ? ` ${reason***REMOVED***` : ''***REMOVED***`;
***REMOVED***

class ResolveError extends Error {***REMOVED***

class InvalidConfigurationError extends ResolveError {
  constructor(context, reason) {
    super(createErrorMsg(context, `Invalid "exports" field. ${reason***REMOVED***`));
  ***REMOVED***
***REMOVED***

class InvalidModuleSpecifierError extends ResolveError {
  constructor(context, internal) {
    super(createErrorMsg(context, internal));
  ***REMOVED***
***REMOVED***

class InvalidPackageTargetError extends ResolveError {
  constructor(context, reason) {
    super(createErrorMsg(context, reason));
  ***REMOVED***
***REMOVED***

/* eslint-disable no-await-in-loop, no-undefined */

function includesInvalidSegments(pathSegments, moduleDirs) {
  return pathSegments
    .split('/')
    .slice(1)
    .some((t) => ['.', '..', ...moduleDirs].includes(t));
***REMOVED***

async function resolvePackageTarget(context, { target, subpath, pattern, internal ***REMOVED***) {
  if (typeof target === 'string') {
    if (!pattern && subpath.length > 0 && !target.endsWith('/')) {
      throw new InvalidModuleSpecifierError(context);
    ***REMOVED***

    if (!target.startsWith('./')) {
      if (internal && !['/', '../'].some((p) => target.startsWith(p)) && !isUrl(target)) {
        // this is a bare package import, remap it and resolve it using regular node resolve
        if (pattern) {
          const result = await context.resolveId(
            target.replace(/\*/g, subpath),
            context.pkgURL.href
          );
          return result ? url.pathToFileURL(result.location) : null;
        ***REMOVED***

        const result = await context.resolveId(`${target***REMOVED***${subpath***REMOVED***`, context.pkgURL.href);
        return result ? url.pathToFileURL(result.location) : null;
      ***REMOVED***
      throw new InvalidPackageTargetError(context, `Invalid mapping: "${target***REMOVED***".`);
    ***REMOVED***

    if (includesInvalidSegments(target, context.moduleDirs)) {
      throw new InvalidPackageTargetError(context, `Invalid mapping: "${target***REMOVED***".`);
    ***REMOVED***

    const resolvedTarget = new URL(target, context.pkgURL);
    if (!resolvedTarget.href.startsWith(context.pkgURL.href)) {
      throw new InvalidPackageTargetError(
        context,
        `Resolved to ${resolvedTarget.href***REMOVED*** which is outside package ${context.pkgURL.href***REMOVED***`
      );
    ***REMOVED***

    if (includesInvalidSegments(subpath, context.moduleDirs)) {
      throw new InvalidModuleSpecifierError(context);
    ***REMOVED***

    if (pattern) {
      return resolvedTarget.href.replace(/\*/g, subpath);
    ***REMOVED***
    return new URL(subpath, resolvedTarget).href;
  ***REMOVED***

  if (Array.isArray(target)) {
    let lastError;
    for (const item of target) {
      try {
        const resolved = await resolvePackageTarget(context, {
          target: item,
          subpath,
          pattern,
          internal
        ***REMOVED***);

        // return if defined or null, but not undefined
        if (resolved !== undefined) {
          return resolved;
        ***REMOVED***
      ***REMOVED*** catch (error) {
        if (!(error instanceof InvalidPackageTargetError)) {
          throw error;
        ***REMOVED*** else {
          lastError = error;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    if (lastError) {
      throw lastError;
    ***REMOVED***
    return null;
  ***REMOVED***

  if (target && typeof target === 'object') {
    for (const [key, value] of Object.entries(target)) {
      if (key === 'default' || context.conditions.includes(key)) {
        const resolved = await resolvePackageTarget(context, {
          target: value,
          subpath,
          pattern,
          internal
        ***REMOVED***);

        // return if defined or null, but not undefined
        if (resolved !== undefined) {
          return resolved;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
    return undefined;
  ***REMOVED***

  if (target === null) {
    return null;
  ***REMOVED***

  throw new InvalidPackageTargetError(context, `Invalid exports field.`);
***REMOVED***

/* eslint-disable no-await-in-loop */

async function resolvePackageImportsExports(context, { matchKey, matchObj, internal ***REMOVED***) {
  if (!matchKey.endsWith('*') && matchKey in matchObj) {
    const target = matchObj[matchKey];
    const resolved = await resolvePackageTarget(context, { target, subpath: '', internal ***REMOVED***);
    return resolved;
  ***REMOVED***

  const expansionKeys = Object.keys(matchObj)
    .filter((k) => k.endsWith('/') || k.endsWith('*'))
    .sort((a, b) => b.length - a.length);

  for (const expansionKey of expansionKeys) {
    const prefix = expansionKey.substring(0, expansionKey.length - 1);

    if (expansionKey.endsWith('*') && matchKey.startsWith(prefix)) {
      const target = matchObj[expansionKey];
      const subpath = matchKey.substring(expansionKey.length - 1);
      const resolved = await resolvePackageTarget(context, {
        target,
        subpath,
        pattern: true,
        internal
      ***REMOVED***);
      return resolved;
    ***REMOVED***

    if (matchKey.startsWith(expansionKey)) {
      const target = matchObj[expansionKey];
      const subpath = matchKey.substring(expansionKey.length);

      const resolved = await resolvePackageTarget(context, { target, subpath, internal ***REMOVED***);
      return resolved;
    ***REMOVED***
  ***REMOVED***

  throw new InvalidModuleSpecifierError(context, internal);
***REMOVED***

async function resolvePackageExports(context, subpath, exports) {
  if (isMixedExports(exports)) {
    throw new InvalidConfigurationError(
      context,
      'All keys must either start with ./, or without one.'
    );
  ***REMOVED***

  if (subpath === '.') {
    let mainExport;
    // If exports is a String or Array, or an Object containing no keys starting with ".", then
    if (typeof exports === 'string' || Array.isArray(exports) || isConditions(exports)) {
      mainExport = exports;
    ***REMOVED*** else if (isMappings(exports)) {
      mainExport = exports['.'];
    ***REMOVED***

    if (mainExport) {
      const resolved = await resolvePackageTarget(context, { target: mainExport, subpath: '' ***REMOVED***);
      if (resolved) {
        return resolved;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED*** else if (isMappings(exports)) {
    const resolvedMatch = await resolvePackageImportsExports(context, {
      matchKey: subpath,
      matchObj: exports
    ***REMOVED***);

    if (resolvedMatch) {
      return resolvedMatch;
    ***REMOVED***
  ***REMOVED***

  throw new InvalidModuleSpecifierError(context);
***REMOVED***

async function resolvePackageImports({
  importSpecifier,
  importer,
  moduleDirs,
  conditions,
  resolveId
***REMOVED***) {
  const result = await findPackageJson(importer, moduleDirs);
  if (!result) {
    throw new Error(createBaseErrorMsg('. Could not find a parent package.json.'));
  ***REMOVED***

  const { pkgPath, pkgJsonPath, pkgJson ***REMOVED*** = result;
  const pkgURL = url.pathToFileURL(`${pkgPath***REMOVED***/`);
  const context = {
    importer,
    importSpecifier,
    moduleDirs,
    pkgURL,
    pkgJsonPath,
    conditions,
    resolveId
  ***REMOVED***;

  const { imports ***REMOVED*** = pkgJson;
  if (!imports) {
    throw new InvalidModuleSpecifierError(context, true);
  ***REMOVED***

  if (importSpecifier === '#' || importSpecifier.startsWith('#/')) {
    throw new InvalidModuleSpecifierError(context, 'Invalid import specifier.');
  ***REMOVED***

  return resolvePackageImportsExports(context, {
    matchKey: importSpecifier,
    matchObj: imports,
    internal: true
  ***REMOVED***);
***REMOVED***

const resolveImportPath = util.promisify(resolve__default['default']);
const readFile$1 = util.promisify(fs__default['default'].readFile);

async function getPackageJson(importer, pkgName, resolveOptions, moduleDirectories) {
  if (importer) {
    const selfPackageJsonResult = await findPackageJson(importer, moduleDirectories);
    if (selfPackageJsonResult && selfPackageJsonResult.pkgJson.name === pkgName) {
      // the referenced package name is the current package
      return selfPackageJsonResult;
    ***REMOVED***
  ***REMOVED***

  try {
    const pkgJsonPath = await resolveImportPath(`${pkgName***REMOVED***/package.json`, resolveOptions);
    const pkgJson = JSON.parse(await readFile$1(pkgJsonPath, 'utf-8'));
    return { pkgJsonPath, pkgJson ***REMOVED***;
  ***REMOVED*** catch (_) {
    return null;
  ***REMOVED***
***REMOVED***

async function resolveId({
  importer,
  importSpecifier,
  exportConditions,
  warn,
  packageInfoCache,
  extensions,
  mainFields,
  preserveSymlinks,
  useBrowserOverrides,
  baseDir,
  moduleDirectories,
  rootDir,
  ignoreSideEffectsForRoot
***REMOVED***) {
  let hasModuleSideEffects = () => null;
  let hasPackageEntry = true;
  let packageBrowserField = false;
  let packageInfo;

  const filter = (pkg, pkgPath) => {
    const info = getPackageInfo({
      cache: packageInfoCache,
      extensions,
      pkg,
      pkgPath,
      mainFields,
      preserveSymlinks,
      useBrowserOverrides,
      rootDir,
      ignoreSideEffectsForRoot
    ***REMOVED***);

    ({ packageInfo, hasModuleSideEffects, hasPackageEntry, packageBrowserField ***REMOVED*** = info);

    return info.cachedPkg;
  ***REMOVED***;

  const resolveOptions = {
    basedir: baseDir,
    readFile: readCachedFile,
    isFile: isFileCached,
    isDirectory: isDirCached,
    extensions,
    includeCoreModules: false,
    moduleDirectory: moduleDirectories,
    preserveSymlinks,
    packageFilter: filter
  ***REMOVED***;

  let location;

  const pkgName = getPackageName(importSpecifier);
  if (importSpecifier.startsWith('#')) {
    // this is a package internal import, resolve using package imports field
    const resolveResult = await resolvePackageImports({
      importSpecifier,
      importer,
      moduleDirs: moduleDirectories,
      conditions: exportConditions,
      resolveId(id, parent) {
        return resolveId({
          importSpecifier: id,
          importer: parent,
          exportConditions,
          warn,
          packageInfoCache,
          extensions,
          mainFields,
          preserveSymlinks,
          useBrowserOverrides,
          baseDir,
          moduleDirectories
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***);
    location = url.fileURLToPath(resolveResult);
  ***REMOVED*** else if (pkgName) {
    // it's a bare import, find the package.json and resolve using package exports if available
    const result = await getPackageJson(importer, pkgName, resolveOptions, moduleDirectories);

    if (result && result.pkgJson.exports) {
      const { pkgJson, pkgJsonPath ***REMOVED*** = result;
      try {
        const subpath =
          pkgName === importSpecifier ? '.' : `.${importSpecifier.substring(pkgName.length)***REMOVED***`;
        const pkgDr = pkgJsonPath.replace('package.json', '');
        const pkgURL = url.pathToFileURL(pkgDr);

        const context = {
          importer,
          importSpecifier,
          moduleDirs: moduleDirectories,
          pkgURL,
          pkgJsonPath,
          conditions: exportConditions
        ***REMOVED***;
        const resolvedPackageExport = await resolvePackageExports(
          context,
          subpath,
          pkgJson.exports
        );
        location = url.fileURLToPath(resolvedPackageExport);
      ***REMOVED*** catch (error) {
        if (error instanceof ResolveError) {
          return error;
        ***REMOVED***
        throw error;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  if (!location) {
    // package has no imports or exports, use classic node resolve
    try {
      location = await resolveImportPath(importSpecifier, resolveOptions);
    ***REMOVED*** catch (error) {
      if (error.code !== 'MODULE_NOT_FOUND') {
        throw error;
      ***REMOVED***
      return null;
    ***REMOVED***
  ***REMOVED***

  if (!preserveSymlinks) {
    if (await exists(location)) {
      location = await realpath(location);
    ***REMOVED***
  ***REMOVED***

  return {
    location,
    hasModuleSideEffects,
    hasPackageEntry,
    packageBrowserField,
    packageInfo
  ***REMOVED***;
***REMOVED***

// Resolve module specifiers in order. Promise resolves to the first module that resolves
// successfully, or the error that resulted from the last attempted module resolution.
async function resolveImportSpecifiers({
  importer,
  importSpecifierList,
  exportConditions,
  warn,
  packageInfoCache,
  extensions,
  mainFields,
  preserveSymlinks,
  useBrowserOverrides,
  baseDir,
  moduleDirectories,
  rootDir,
  ignoreSideEffectsForRoot
***REMOVED***) {
  let lastResolveError;

  for (let i = 0; i < importSpecifierList.length; i++) {
    // eslint-disable-next-line no-await-in-loop
    const result = await resolveId({
      importer,
      importSpecifier: importSpecifierList[i],
      exportConditions,
      warn,
      packageInfoCache,
      extensions,
      mainFields,
      preserveSymlinks,
      useBrowserOverrides,
      baseDir,
      moduleDirectories,
      rootDir,
      ignoreSideEffectsForRoot
    ***REMOVED***);

    if (result instanceof ResolveError) {
      lastResolveError = result;
    ***REMOVED*** else if (result) {
      return result;
    ***REMOVED***
  ***REMOVED***

  if (lastResolveError) {
    // only log the last failed resolve error
    warn(lastResolveError);
  ***REMOVED***
  return null;
***REMOVED***

function handleDeprecatedOptions(opts) {
  const warnings = [];

  if (opts.customResolveOptions) {
    const { customResolveOptions ***REMOVED*** = opts;
    if (customResolveOptions.moduleDirectory) {
      // eslint-disable-next-line no-param-reassign
      opts.moduleDirectories = Array.isArray(customResolveOptions.moduleDirectory)
        ? customResolveOptions.moduleDirectory
        : [customResolveOptions.moduleDirectory];

      warnings.push(
        'node-resolve: The `customResolveOptions.moduleDirectory` option has been deprecated. Use `moduleDirectories`, which must be an array.'
      );
    ***REMOVED***

    if (customResolveOptions.preserveSymlinks) {
      throw new Error(
        'node-resolve: `customResolveOptions.preserveSymlinks` is no longer an option. We now always use the rollup `preserveSymlinks` option.'
      );
    ***REMOVED***

    [
      'basedir',
      'package',
      'extensions',
      'includeCoreModules',
      'readFile',
      'isFile',
      'isDirectory',
      'realpath',
      'packageFilter',
      'pathFilter',
      'paths',
      'packageIterator'
    ].forEach((resolveOption) => {
      if (customResolveOptions[resolveOption]) {
        throw new Error(
          `node-resolve: \`customResolveOptions.${resolveOption***REMOVED***\` is no longer an option. If you need this, please open an issue.`
        );
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***

  return { warnings ***REMOVED***;
***REMOVED***

/* eslint-disable no-param-reassign, no-shadow, no-undefined */

const builtins = new Set(builtinList__default['default']);
const ES6_BROWSER_EMPTY = '\0node-resolve:empty.js';
const deepFreeze = (object) => {
  Object.freeze(object);

  for (const value of Object.values(object)) {
    if (typeof value === 'object' && !Object.isFrozen(value)) {
      deepFreeze(value);
    ***REMOVED***
  ***REMOVED***

  return object;
***REMOVED***;

const baseConditions = ['default', 'module'];
const baseConditionsEsm = [...baseConditions, 'import'];
const baseConditionsCjs = [...baseConditions, 'require'];
const defaults = {
  dedupe: [],
  // It's important that .mjs is listed before .js so that Rollup will interpret npm modules
  // which deploy both ESM .mjs and CommonJS .js files as ESM.
  extensions: ['.mjs', '.js', '.json', '.node'],
  resolveOnly: [],
  moduleDirectories: ['node_modules'],
  ignoreSideEffectsForRoot: false
***REMOVED***;
const DEFAULTS = deepFreeze(deepMerge__default['default']({***REMOVED***, defaults));

function nodeResolve(opts = {***REMOVED***) {
  const { warnings ***REMOVED*** = handleDeprecatedOptions(opts);

  const options = { ...defaults, ...opts ***REMOVED***;
  const { extensions, jail, moduleDirectories, ignoreSideEffectsForRoot ***REMOVED*** = options;
  const conditionsEsm = [...baseConditionsEsm, ...(options.exportConditions || [])];
  const conditionsCjs = [...baseConditionsCjs, ...(options.exportConditions || [])];
  const packageInfoCache = new Map();
  const idToPackageInfo = new Map();
  const mainFields = getMainFields(options);
  const useBrowserOverrides = mainFields.indexOf('browser') !== -1;
  const isPreferBuiltinsSet = options.preferBuiltins === true || options.preferBuiltins === false;
  const preferBuiltins = isPreferBuiltinsSet ? options.preferBuiltins : true;
  const rootDir = path.resolve(options.rootDir || process.cwd());
  let { dedupe ***REMOVED*** = options;
  let rollupOptions;

  if (typeof dedupe !== 'function') {
    dedupe = (importee) =>
      options.dedupe.includes(importee) || options.dedupe.includes(getPackageName(importee));
  ***REMOVED***

  const resolveOnly = options.resolveOnly.map((pattern) => {
    if (pattern instanceof RegExp) {
      return pattern;
    ***REMOVED***
    const normalized = pattern.replace(/[\\^$*+?.()|[\]{***REMOVED***]/g, '\\$&');
    return new RegExp(`^${normalized***REMOVED***$`);
  ***REMOVED***);

  const browserMapCache = new Map();
  let preserveSymlinks;

  return {
    name: 'node-resolve',

    buildStart(options) {
      rollupOptions = options;

      for (const warning of warnings) {
        this.warn(warning);
      ***REMOVED***

      ({ preserveSymlinks ***REMOVED*** = options);
    ***REMOVED***,

    generateBundle() {
      readCachedFile.clear();
      isFileCached.clear();
      isDirCached.clear();
    ***REMOVED***,

    async resolveId(importee, importer, opts) {
      if (importee === ES6_BROWSER_EMPTY) {
        return importee;
      ***REMOVED***
      // ignore IDs with null character, these belong to other plugins
      if (/\0/.test(importee)) return null;

      if (/\0/.test(importer)) {
        importer = undefined;
      ***REMOVED***

      // strip query params from import
      const [importPath, params] = importee.split('?');
      const importSuffix = `${params ? `?${params***REMOVED***` : ''***REMOVED***`;
      importee = importPath;

      const baseDir = !importer || dedupe(importee) ? rootDir : path.dirname(importer);

      // https://github.com/defunctzombie/package-browser-field-spec
      const browser = browserMapCache.get(importer);
      if (useBrowserOverrides && browser) {
        const resolvedImportee = path.resolve(baseDir, importee);
        if (browser[importee] === false || browser[resolvedImportee] === false) {
          return ES6_BROWSER_EMPTY;
        ***REMOVED***
        const browserImportee =
          browser[importee] ||
          browser[resolvedImportee] ||
          browser[`${resolvedImportee***REMOVED***.js`] ||
          browser[`${resolvedImportee***REMOVED***.json`];
        if (browserImportee) {
          importee = browserImportee;
        ***REMOVED***
      ***REMOVED***

      const parts = importee.split(/[/\\]/);
      let id = parts.shift();
      let isRelativeImport = false;

      if (id[0] === '@' && parts.length > 0) {
        // scoped packages
        id += `/${parts.shift()***REMOVED***`;
      ***REMOVED*** else if (id[0] === '.') {
        // an import relative to the parent dir of the importer
        id = path.resolve(baseDir, importee);
        isRelativeImport = true;
      ***REMOVED***

      if (
        !isRelativeImport &&
        resolveOnly.length &&
        !resolveOnly.some((pattern) => pattern.test(id))
      ) {
        if (normalizeInput(rollupOptions.input).includes(importee)) {
          return null;
        ***REMOVED***
        return false;
      ***REMOVED***

      const importSpecifierList = [];

      if (importer === undefined && !importee[0].match(/^\.?\.?\//)) {
        // For module graph roots (i.e. when importer is undefined), we
        // need to handle 'path fragments` like `foo/bar` that are commonly
        // found in rollup config files. If importee doesn't look like a
        // relative or absolute path, we make it relative and attempt to
        // resolve it. If we don't find anything, we try resolving it as we
        // got it.
        importSpecifierList.push(`./${importee***REMOVED***`);
      ***REMOVED***

      const importeeIsBuiltin = builtins.has(importee);

      if (importeeIsBuiltin) {
        // The `resolve` library will not resolve packages with the same
        // name as a node built-in module. If we're resolving something
        // that's a builtin, and we don't prefer to find built-ins, we
        // first try to look up a local module with that name. If we don't
        // find anything, we resolve the builtin which just returns back
        // the built-in's name.
        importSpecifierList.push(`${importee***REMOVED***/`);
      ***REMOVED***

      // TypeScript files may import '.js' to refer to either '.ts' or '.tsx'
      if (importer && importee.endsWith('.js')) {
        for (const ext of ['.ts', '.tsx']) {
          if (importer.endsWith(ext) && extensions.includes(ext)) {
            importSpecifierList.push(importee.replace(/.js$/, ext));
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***

      importSpecifierList.push(importee);

      const warn = (...args) => this.warn(...args);
      const isRequire =
        opts && opts.custom && opts.custom['node-resolve'] && opts.custom['node-resolve'].isRequire;
      const exportConditions = isRequire ? conditionsCjs : conditionsEsm;

      const resolvedWithoutBuiltins = await resolveImportSpecifiers({
        importer,
        importSpecifierList,
        exportConditions,
        warn,
        packageInfoCache,
        extensions,
        mainFields,
        preserveSymlinks,
        useBrowserOverrides,
        baseDir,
        moduleDirectories,
        rootDir,
        ignoreSideEffectsForRoot
      ***REMOVED***);

      const resolved =
        importeeIsBuiltin && preferBuiltins
          ? {
              packageInfo: undefined,
              hasModuleSideEffects: () => null,
              hasPackageEntry: true,
              packageBrowserField: false
            ***REMOVED***
          : resolvedWithoutBuiltins;
      if (!resolved) {
        return null;
      ***REMOVED***

      const { packageInfo, hasModuleSideEffects, hasPackageEntry, packageBrowserField ***REMOVED*** = resolved;
      let { location ***REMOVED*** = resolved;
      if (packageBrowserField) {
        if (Object.prototype.hasOwnProperty.call(packageBrowserField, location)) {
          if (!packageBrowserField[location]) {
            browserMapCache.set(location, packageBrowserField);
            return ES6_BROWSER_EMPTY;
          ***REMOVED***
          location = packageBrowserField[location];
        ***REMOVED***
        browserMapCache.set(location, packageBrowserField);
      ***REMOVED***

      if (hasPackageEntry && !preserveSymlinks) {
        const fileExists = await exists(location);
        if (fileExists) {
          location = await realpath(location);
        ***REMOVED***
      ***REMOVED***

      idToPackageInfo.set(location, packageInfo);

      if (hasPackageEntry) {
        if (importeeIsBuiltin && preferBuiltins) {
          if (!isPreferBuiltinsSet && resolvedWithoutBuiltins && resolved !== importee) {
            this.warn(
              `preferring built-in module '${importee***REMOVED***' over local alternative at '${resolvedWithoutBuiltins.location***REMOVED***', pass 'preferBuiltins: false' to disable this behavior or 'preferBuiltins: true' to disable this warning`
            );
          ***REMOVED***
          return false;
        ***REMOVED*** else if (jail && location.indexOf(path.normalize(jail.trim(path.sep))) !== 0) {
          return null;
        ***REMOVED***
      ***REMOVED***

      if (options.modulesOnly && (await exists(location))) {
        const code = await readFile(location, 'utf-8');
        if (isModule__default['default'](code)) {
          return {
            id: `${location***REMOVED***${importSuffix***REMOVED***`,
            moduleSideEffects: hasModuleSideEffects(location)
          ***REMOVED***;
        ***REMOVED***
        return null;
      ***REMOVED***
      const result = {
        id: `${location***REMOVED***${importSuffix***REMOVED***`,
        moduleSideEffects: hasModuleSideEffects(location)
      ***REMOVED***;
      return result;
    ***REMOVED***,

    load(importee) {
      if (importee === ES6_BROWSER_EMPTY) {
        return 'export default {***REMOVED***;';
      ***REMOVED***
      return null;
    ***REMOVED***,

    getPackageInfoForId(id) {
      return idToPackageInfo.get(id);
    ***REMOVED***
  ***REMOVED***;
***REMOVED***

exports.DEFAULTS = DEFAULTS;
exports.default = nodeResolve;
exports.nodeResolve = nodeResolve;
