// jest-environment-node/build/index.js
'use strict';

function _vm() {
  const data = require('vm');

  _vm = function () {
    return data;
  ***REMOVED***;

  return data;
***REMOVED***

function _fakeTimers() {
  const data = require('@jest/fake-timers');

  _fakeTimers = function () {
    return data;
  ***REMOVED***;

  return data;
***REMOVED***

function _jestMock() {
  const data = require('jest-mock');

  _jestMock = function () {
    return data;
  ***REMOVED***;

  return data;
***REMOVED***

function _jestUtil() {
  const data = require('jest-util');

  _jestUtil = function () {
    return data;
  ***REMOVED***;

  return data;
***REMOVED***

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    ***REMOVED***);
  ***REMOVED*** else {
    obj[key] = value;
  ***REMOVED***
  return obj;
***REMOVED***

class NodeEnvironment {
  constructor(config) {
    _defineProperty(this, 'context', void 0);

    _defineProperty(this, 'fakeTimers', void 0);

    _defineProperty(this, 'fakeTimersModern', void 0);

    _defineProperty(this, 'global', void 0);

    _defineProperty(this, 'moduleMocker', void 0);

    this.context = (0, _vm().createContext)();
    const global = (this.global = (0, _vm().runInContext)(
      'this',
      Object.assign(this.context, config.testEnvironmentOptions)
    ));
    global.global = global;
    global.clearInterval = clearInterval;
    global.clearTimeout = clearTimeout;
    global.setInterval = setInterval;
    global.setTimeout = setTimeout;
    global.Buffer = Buffer;
    global.setImmediate = setImmediate;
    global.clearImmediate = clearImmediate;
    global.ArrayBuffer = ArrayBuffer; // TextEncoder (global or via 'util') references a Uint8Array constructor
    // different than the global one used by users in tests. This makes sure the
    // same constructor is referenced by both.

    global.Uint8Array = Uint8Array; // URL and URLSearchParams are global in Node >= 10

    if (typeof URL !== 'undefined' && typeof URLSearchParams !== 'undefined') {
      global.URL = URL;
      global.URLSearchParams = URLSearchParams;
    ***REMOVED*** // TextDecoder and TextDecoder are global in Node >= 11

    if (
      typeof TextEncoder !== 'undefined' &&
      typeof TextDecoder !== 'undefined'
    ) {
      global.TextEncoder = TextEncoder;
      global.TextDecoder = TextDecoder;
    ***REMOVED*** // queueMicrotask is global in Node >= 11

    if (typeof queueMicrotask !== 'undefined') {
      global.queueMicrotask = queueMicrotask;
    ***REMOVED*** // AbortController is global in Node >= 15

    if (typeof AbortController !== 'undefined') {
      global.AbortController = AbortController;
    ***REMOVED*** // AbortSignal is global in Node >= 15

    if (typeof AbortSignal !== 'undefined') {
      global.AbortSignal = AbortSignal;
    ***REMOVED*** // Event is global in Node >= 15.4

    if (typeof Event !== 'undefined') {
      global.Event = Event;
    ***REMOVED*** // EventTarget is global in Node >= 15.4

    if (typeof EventTarget !== 'undefined') {
      global.EventTarget = EventTarget;
    ***REMOVED*** // performance is global in Node >= 16

    if (typeof performance !== 'undefined') {
      global.performance = performance;
    ***REMOVED*** // atob and btoa are global in Node >= 16

    if (typeof atob !== 'undefined' && typeof btoa !== 'undefined') {
      global.atob = atob;
      global.btoa = btoa;
    ***REMOVED***

    (0, _jestUtil().installCommonGlobals)(global, config.globals);
    this.moduleMocker = new (_jestMock().ModuleMocker)(global);

    const timerIdToRef = id => ({
      id,

      ref() {
        return this;
      ***REMOVED***,

      unref() {
        return this;
      ***REMOVED***
    ***REMOVED***);

    const timerRefToId = timer => (timer && timer.id) || undefined;

    const timerConfig = {
      idToRef: timerIdToRef,
      refToId: timerRefToId
    ***REMOVED***;
    this.fakeTimers = new (_fakeTimers().LegacyFakeTimers)({
      config,
      global,
      moduleMocker: this.moduleMocker,
      timerConfig
    ***REMOVED***);
    this.fakeTimersModern = new (_fakeTimers().ModernFakeTimers)({
      config,
      global
    ***REMOVED***);
  ***REMOVED***

  async setup() {***REMOVED***

  async teardown() {
    if (this.fakeTimers) {
      this.fakeTimers.dispose();
    ***REMOVED***

    if (this.fakeTimersModern) {
      this.fakeTimersModern.dispose();
    ***REMOVED***

    this.context = null;
    this.fakeTimers = null;
    this.fakeTimersModern = null;
  ***REMOVED***

  getVmContext() {
    return this.context;
  ***REMOVED***
***REMOVED***

module.exports = NodeEnvironment;
