// filelist/index.js
/*
 * Jake JavaScript build tool
 * Copyright 2112 Matthew Eernisse (mde@fleegix.org)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
*/
var fs = require('fs')
, path = require('path')
, minimatch = require('minimatch')
, escapeRegExpChars
, merge
, basedir
, _readDir
, readdirR
, globSync;
var hasOwnProperty = Object.prototype.hasOwnProperty;
var hasOwn = function (obj, key) { return hasOwnProperty.apply(obj, [key]); ***REMOVED***;

  /**
    @name escapeRegExpChars
    @function
    @return {String***REMOVED*** A string of escaped characters
    @description Escapes regex control-characters in strings
                 used to build regexes dynamically
    @param {String***REMOVED*** string The string of chars to escape
  */
  escapeRegExpChars = (function () {
    var specials = [ '^', '$', '/', '.', '*', '+', '?', '|', '(', ')',
        '[', ']', '{', '***REMOVED***', '\\' ];
    var sRE = new RegExp('(\\' + specials.join('|\\') + ')', 'g');
    return function (string) {
      var str = string || '';
      str = String(str);
      return str.replace(sRE, '\\$1');
    ***REMOVED***;
  ***REMOVED***)();

  /**
    @name merge
    @function
    @return {Object***REMOVED*** Returns the merged object
    @description Merge merges `otherObject` into `object` and takes care of deep
                 merging of objects
    @param {Object***REMOVED*** object Object to merge into
    @param {Object***REMOVED*** otherObject Object to read from
  */
  merge = function (object, otherObject) {
    var obj = object || {***REMOVED***
      , otherObj = otherObject || {***REMOVED***
      , key, value;

    for (key in otherObj) {

      if (!hasOwn(otherObj, key)) {
        continue;
      ***REMOVED***
      if (key === '__proto__' || key === 'constructor') {
        continue;
      ***REMOVED***

      value = otherObj[key];
      // Check if a value is an Object, if so recursively add it's key/values
      if (typeof value === 'object' && !(value instanceof Array)) {
        // Update value of object to the one from otherObj
        obj[key] = merge(obj[key], value);
      ***REMOVED***
      // Value is anything other than an Object, so just add it
      else {
        obj[key] = value;
      ***REMOVED***
    ***REMOVED***

    return obj;
  ***REMOVED***;
  /**
    Given a patern, return the base directory of it (ie. the folder
    that will contain all the files matching the path).
    eg. file.basedir('/test/**') => '/test/'
    Path ending by '/' are considerd as folder while other are considerd
    as files, eg.:
        file.basedir('/test/a/') => '/test/a'
        file.basedir('/test/a') => '/test'
    The returned path always end with a '/' so we have:
        file.basedir(file.basedir(x)) == file.basedir(x)
  */
  basedir = function (pathParam) {
    var bd = ''
      , parts
      , part
      , pos = 0
      , p = pathParam || '';

    // If the path has a leading asterisk, basedir is the current dir
    if (p.indexOf('*') == 0 || p.indexOf('**') == 0) {
      return '.';
    ***REMOVED***

    // always consider .. at the end as a folder and not a filename
    if (/(?:^|\/|\\)\.\.$/.test(p.slice(-3))) {
      p += '/';
    ***REMOVED***

    parts = p.split(/\\|\//);
    for (var i = 0, l = parts.length - 1; i < l; i++) {
      part = parts[i];
      if (part.indexOf('*') > -1 || part.indexOf('**') > -1) {
        break;
      ***REMOVED***
      pos += part.length + 1;
      bd += part + p[pos - 1];
    ***REMOVED***
    if (!bd) {
      bd = '.';
    ***REMOVED***
    // Strip trailing slashes
    if (!(bd == '\\' || bd == '/')) {
      bd = bd.replace(/\\$|\/$/, '');
    ***REMOVED***
    return bd;

  ***REMOVED***;

  // Return the contents of a given directory
  _readDir = function (dirPath) {
    var dir = path.normalize(dirPath)
      , paths = []
      , ret = [dir]
      , msg;

    try {
      paths = fs.readdirSync(dir);
    ***REMOVED***
    catch (e) {
      msg = 'Could not read path ' + dir + '\n';
      if (e.stack) {
        msg += e.stack;
      ***REMOVED***
      throw new Error(msg);
    ***REMOVED***

    paths.forEach(function (p) {
      var curr = path.join(dir, p);
      var stat = fs.statSync(curr);
      if (stat.isDirectory()) {
        ret = ret.concat(_readDir(curr));
      ***REMOVED***
      else {
        ret.push(curr);
      ***REMOVED***
    ***REMOVED***);

    return ret;
  ***REMOVED***;

  /**
    @name file#readdirR
    @function
    @return {Array***REMOVED*** Returns the contents as an Array, can be configured via opts.format
    @description Reads the given directory returning it's contents
    @param {String***REMOVED*** dir The directory to read
    @param {Object***REMOVED*** opts Options to use
      @param {String***REMOVED*** [opts.format] Set the format to return(Default: Array)
  */
  readdirR = function (dir, opts) {
    var options = opts || {***REMOVED***
      , format = options.format || 'array'
      , ret;
    ret = _readDir(dir);
    return format == 'string' ? ret.join('\n') : ret;
  ***REMOVED***;


globSync = function (pat, opts) {
  var dirname = basedir(pat)
    , files
    , matches;

  try {
    files = readdirR(dirname).map(function(file){
      return file.replace(/\\/g, '/');
    ***REMOVED***);
  ***REMOVED***
  // Bail if path doesn't exist -- assume no files
  catch(e) {
    if (FileList.verbose) console.error(e.message);
  ***REMOVED***

  if (files) {
    pat = path.normalize(pat);
    matches = minimatch.match(files, pat, opts || {***REMOVED***);
  ***REMOVED***
  return matches || [];
***REMOVED***;

// Constants
// ---------------
// List of all the builtin Array methods we want to override
var ARRAY_METHODS = Object.getOwnPropertyNames(Array.prototype)
// Array methods that return a copy instead of affecting the original
  , SPECIAL_RETURN = {
      'concat': true
    , 'slice': true
    , 'filter': true
    , 'map': true
    ***REMOVED***
// Default file-patterns we want to ignore
  , DEFAULT_IGNORE_PATTERNS = [
      /(^|[\/\\])CVS([\/\\]|$)/
    , /(^|[\/\\])\.svn([\/\\]|$)/
    , /(^|[\/\\])\.git([\/\\]|$)/
    , /\.bak$/
    , /~$/
    ]
// Ignore core files
  , DEFAULT_IGNORE_FUNCS = [
      function (name) {
        var isDir = false
          , stats;
        try {
          stats = fs.statSync(name);
          isDir = stats.isDirectory();
        ***REMOVED***
        catch(e) {***REMOVED***
        return (/(^|[\/\\])core$/).test(name) && !isDir;
      ***REMOVED***
    ];

var FileList = function () {
  var self = this
    , wrap;

  // List of glob-patterns or specific filenames
  this.pendingAdd = [];
  // Switched to false after lazy-eval of files
  this.pending = true;
  // Used to calculate exclusions from the list of files
  this.excludes = {
    pats: DEFAULT_IGNORE_PATTERNS.slice()
  , funcs: DEFAULT_IGNORE_FUNCS.slice()
  , regex: null
  ***REMOVED***;
  this.items = [];

  // Wrap the array methods with the delegates
  wrap = function (prop) {
    var arr;
    self[prop] = function () {
      if (self.pending) {
        self.resolve();
      ***REMOVED***
      if (typeof self.items[prop] == 'function') {
        // Special method that return a copy
        if (SPECIAL_RETURN[prop]) {
          arr = self.items[prop].apply(self.items, arguments);
          return FileList.clone(self, arr);
        ***REMOVED***
        else {
          return self.items[prop].apply(self.items, arguments);
        ***REMOVED***
      ***REMOVED***
      else {
        return self.items[prop];
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***;
  for (var i = 0, ii = ARRAY_METHODS.length; i < ii; i++) {
    wrap(ARRAY_METHODS[i]);
  ***REMOVED***

  // Include whatever files got passed to the constructor
  this.include.apply(this, arguments);

  // Fix constructor linkage
  this.constructor = FileList;
***REMOVED***;

FileList.prototype = new (function () {
  var globPattern = /[*?\[\{]/;

  var _addMatching = function (item) {
        var matches = globSync(item.path, item.options);
        this.items = this.items.concat(matches);
      ***REMOVED***

    , _resolveAdd = function (item) {
        if (globPattern.test(item.path)) {
          _addMatching.call(this, item);
        ***REMOVED***
        else {
          this.push(item.path);
        ***REMOVED***
      ***REMOVED***

    , _calculateExcludeRe = function () {
        var pats = this.excludes.pats
          , pat
          , excl = []
          , matches = [];

        for (var i = 0, ii = pats.length; i < ii; i++) {
          pat = pats[i];
          if (typeof pat == 'string') {
            // Glob, look up files
            if (/[*?]/.test(pat)) {
              matches = globSync(pat);
              matches = matches.map(function (m) {
                return escapeRegExpChars(m);
              ***REMOVED***);
              excl = excl.concat(matches);
            ***REMOVED***
            // String for regex
            else {
              excl.push(escapeRegExpChars(pat));
            ***REMOVED***
          ***REMOVED***
          // Regex, grab the string-representation
          else if (pat instanceof RegExp) {
            excl.push(pat.toString().replace(/^\/|\/$/g, ''));
          ***REMOVED***
        ***REMOVED***
        if (excl.length) {
          this.excludes.regex = new RegExp('(' + excl.join(')|(') + ')');
        ***REMOVED***
        else {
          this.excludes.regex = /^$/;
        ***REMOVED***
      ***REMOVED***

    , _resolveExclude = function () {
        var self = this;
        _calculateExcludeRe.call(this);
        // No `reject` method, so use reverse-filter
        this.items = this.items.filter(function (name) {
          return !self.shouldExclude(name);
        ***REMOVED***);
      ***REMOVED***;

  /**
   * Includes file-patterns in the FileList. Should be called with one or more
   * pattern for finding file to include in the list. Arguments should be strings
   * for either a glob-pattern or a specific file-name, or an array of them
   */
  this.include = function () {
    var args = Array.prototype.slice.call(arguments)
        , arg
        , includes = { items: [], options: {***REMOVED*** ***REMOVED***;

    for (var i = 0, ilen = args.length; i < ilen; i++) {
      arg = args[i];

      if (typeof arg === 'object' && !Array.isArray(arg)) {
        merge(includes.options, arg);
      ***REMOVED*** else {
        includes.items = includes.items.concat(arg).filter(function (item) {
          return !!item;
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***

    var items = includes.items.map(function(item) {
      return { path: item, options: includes.options ***REMOVED***;
    ***REMOVED***);

    this.pendingAdd = this.pendingAdd.concat(items);

    return this;
  ***REMOVED***;

  /**
   * Indicates whether a particular file would be filtered out by the current
   * exclusion rules for this FileList.
   * @param {String***REMOVED*** name The filename to check
   * @return {Boolean***REMOVED*** Whether or not the file should be excluded
   */
  this.shouldExclude = function (name) {
    if (!this.excludes.regex) {
      _calculateExcludeRe.call(this);
    ***REMOVED***
    var excl = this.excludes;
    return excl.regex.test(name) || excl.funcs.some(function (f) {
      return !!f(name);
    ***REMOVED***);
  ***REMOVED***;

  /**
   * Excludes file-patterns from the FileList. Should be called with one or more
   * pattern for finding file to include in the list. Arguments can be:
   * 1. Strings for either a glob-pattern or a specific file-name
   * 2. Regular expression literals
   * 3. Functions to be run on the filename that return a true/false
   */
  this.exclude = function () {
    var args = Array.isArray(arguments[0]) ? arguments[0] : arguments
      , arg;
    for (var i = 0, ii = args.length; i < ii; i++) {
      arg = args[i];
      if (typeof arg == 'function' && !(arg instanceof RegExp)) {
        this.excludes.funcs.push(arg);
      ***REMOVED***
      else {
        this.excludes.pats.push(arg);
      ***REMOVED***
    ***REMOVED***
    if (!this.pending) {
      _resolveExclude.call(this);
    ***REMOVED***
    return this;
  ***REMOVED***;

  /**
   * Populates the FileList from the include/exclude rules with a list of
   * actual files
   */
  this.resolve = function () {
    var item
      , uniqueFunc = function (p, c) {
          if (p.indexOf(c) < 0) {
            p.push(c);
          ***REMOVED***
          return p;
        ***REMOVED***;
    if (this.pending) {
      this.pending = false;
      while ((item = this.pendingAdd.shift())) {
        _resolveAdd.call(this, item);
      ***REMOVED***
      // Reduce to a unique list
      this.items = this.items.reduce(uniqueFunc, []);
      // Remove exclusions
      _resolveExclude.call(this);
    ***REMOVED***
    return this;
  ***REMOVED***;

  /**
   * Convert to a plain-jane array
   */
  this.toArray = function () {
    // Call slice to ensure lazy-resolution before slicing items
    var ret = this.slice().items.slice();
    return ret;
  ***REMOVED***;

  /**
   * Clear any pending items -- only useful before
   * calling `resolve`
   */
  this.clearInclusions = function () {
    this.pendingAdd = [];
    return this;
  ***REMOVED***;

  /**
   * Clear any current exclusion rules
   */
  this.clearExclusions = function () {
    this.excludes = {
      pats: []
    , funcs: []
    , regex: null
    ***REMOVED***;
    return this;
  ***REMOVED***;

***REMOVED***)();

// Static method, used to create copy returned by special
// array methods
FileList.clone = function (list, items) {
  var clone = new FileList();
  if (items) {
    clone.items = items;
  ***REMOVED***
  clone.pendingAdd = list.pendingAdd;
  clone.pending = list.pending;
  for (var p in list.excludes) {
    clone.excludes[p] = list.excludes[p];
  ***REMOVED***
  return clone;
***REMOVED***;

FileList.verbose = true

exports.FileList = FileList;
