// fast-glob/out/readers/reader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const path = require("path");
const fsStat = require("@nodelib/fs.stat");
const utils = require("../utils");
class Reader {
    constructor(_settings) {
        this._settings = _settings;
        this._fsStatSettings = new fsStat.Settings({
            followSymbolicLink: this._settings.followSymbolicLinks,
            fs: this._settings.fs,
            throwErrorOnBrokenSymbolicLink: this._settings.followSymbolicLinks
        ***REMOVED***);
    ***REMOVED***
    _getFullEntryPath(filepath) {
        return path.resolve(this._settings.cwd, filepath);
    ***REMOVED***
    _makeEntry(stats, pattern) {
        const entry = {
            name: pattern,
            path: pattern,
            dirent: utils.fs.createDirentFromStats(pattern, stats)
        ***REMOVED***;
        if (this._settings.stats) {
            entry.stats = stats;
        ***REMOVED***
        return entry;
    ***REMOVED***
    _isFatalError(error) {
        return !utils.errno.isEnoentCodeError(error) && !this._settings.suppressErrors;
    ***REMOVED***
***REMOVED***
exports.default = Reader;
