// object-hash/index.js
'use strict';

var crypto = require('crypto');

/**
 * Exported function
 *
 * Options:
 *
 *  - `algorithm` hash algo to be used by this instance: *'sha1', 'md5'
 *  - `excludeValues` {true|*false***REMOVED*** hash object keys, values ignored
 *  - `encoding` hash encoding, supports 'buffer', '*hex', 'binary', 'base64'
 *  - `ignoreUnknown` {true|*false***REMOVED*** ignore unknown object types
 *  - `replacer` optional function that replaces values before hashing
 *  - `respectFunctionProperties` {*true|false***REMOVED*** consider function properties when hashing
 *  - `respectFunctionNames` {*true|false***REMOVED*** consider 'name' property of functions for hashing
 *  - `respectType` {*true|false***REMOVED*** Respect special properties (prototype, constructor)
 *    when hashing to distinguish between types
 *  - `unorderedArrays` {true|*false***REMOVED*** Sort all arrays before hashing
 *  - `unorderedSets` {*true|false***REMOVED*** Sort `Set` and `Map` instances before hashing
 *  * = default
 *
 * @param {object***REMOVED*** object value to hash
 * @param {object***REMOVED*** options hashing options
 * @return {string***REMOVED*** hash value
 * @api public
 */
exports = module.exports = objectHash;

function objectHash(object, options){
  options = applyDefaults(object, options);

  return hash(object, options);
***REMOVED***

/**
 * Exported sugar methods
 *
 * @param {object***REMOVED*** object value to hash
 * @return {string***REMOVED*** hash value
 * @api public
 */
exports.sha1 = function(object){
  return objectHash(object);
***REMOVED***;
exports.keys = function(object){
  return objectHash(object, {excludeValues: true, algorithm: 'sha1', encoding: 'hex'***REMOVED***);
***REMOVED***;
exports.MD5 = function(object){
  return objectHash(object, {algorithm: 'md5', encoding: 'hex'***REMOVED***);
***REMOVED***;
exports.keysMD5 = function(object){
  return objectHash(object, {algorithm: 'md5', encoding: 'hex', excludeValues: true***REMOVED***);
***REMOVED***;

// Internals
var hashes = crypto.getHashes ? crypto.getHashes().slice() : ['sha1', 'md5'];
hashes.push('passthrough');
var encodings = ['buffer', 'hex', 'binary', 'base64'];

function applyDefaults(object, sourceOptions){
  sourceOptions = sourceOptions || {***REMOVED***;

  // create a copy rather than mutating
  var options = {***REMOVED***;
  options.algorithm = sourceOptions.algorithm || 'sha1';
  options.encoding = sourceOptions.encoding || 'hex';
  options.excludeValues = sourceOptions.excludeValues ? true : false;
  options.algorithm = options.algorithm.toLowerCase();
  options.encoding = options.encoding.toLowerCase();
  options.ignoreUnknown = sourceOptions.ignoreUnknown !== true ? false : true; // default to false
  options.respectType = sourceOptions.respectType === false ? false : true; // default to true
  options.respectFunctionNames = sourceOptions.respectFunctionNames === false ? false : true;
  options.respectFunctionProperties = sourceOptions.respectFunctionProperties === false ? false : true;
  options.unorderedArrays = sourceOptions.unorderedArrays !== true ? false : true; // default to false
  options.unorderedSets = sourceOptions.unorderedSets === false ? false : true; // default to false
  options.unorderedObjects = sourceOptions.unorderedObjects === false ? false : true; // default to true
  options.replacer = sourceOptions.replacer || undefined;
  options.excludeKeys = sourceOptions.excludeKeys || undefined;

  if(typeof object === 'undefined') {
    throw new Error('Object argument required.');
  ***REMOVED***

  // if there is a case-insensitive match in the hashes list, accept it
  // (i.e. SHA256 for sha256)
  for (var i = 0; i < hashes.length; ++i) {
    if (hashes[i].toLowerCase() === options.algorithm.toLowerCase()) {
      options.algorithm = hashes[i];
    ***REMOVED***
  ***REMOVED***

  if(hashes.indexOf(options.algorithm) === -1){
    throw new Error('Algorithm "' + options.algorithm + '"  not supported. ' +
      'supported values: ' + hashes.join(', '));
  ***REMOVED***

  if(encodings.indexOf(options.encoding) === -1 &&
     options.algorithm !== 'passthrough'){
    throw new Error('Encoding "' + options.encoding + '"  not supported. ' +
      'supported values: ' + encodings.join(', '));
  ***REMOVED***

  return options;
***REMOVED***

/** Check if the given function is a native function */
function isNativeFunction(f) {
  if ((typeof f) !== 'function') {
    return false;
  ***REMOVED***
  var exp = /^function\s+\w*\s*\(\s*\)\s*{\s+\[native code\]\s+***REMOVED***$/i;
  return exp.exec(Function.prototype.toString.call(f)) != null;
***REMOVED***

function hash(object, options) {
  var hashingStream;

  if (options.algorithm !== 'passthrough') {
    hashingStream = crypto.createHash(options.algorithm);
  ***REMOVED*** else {
    hashingStream = new PassThrough();
  ***REMOVED***

  if (typeof hashingStream.write === 'undefined') {
    hashingStream.write = hashingStream.update;
    hashingStream.end   = hashingStream.update;
  ***REMOVED***

  var hasher = typeHasher(options, hashingStream);
  hasher.dispatch(object);
  if (!hashingStream.update) {
    hashingStream.end('');
  ***REMOVED***

  if (hashingStream.digest) {
    return hashingStream.digest(options.encoding === 'buffer' ? undefined : options.encoding);
  ***REMOVED***

  var buf = hashingStream.read();
  if (options.encoding === 'buffer') {
    return buf;
  ***REMOVED***

  return buf.toString(options.encoding);
***REMOVED***

/**
 * Expose streaming API
 *
 * @param {object***REMOVED*** object  Value to serialize
 * @param {object***REMOVED*** options  Options, as for hash()
 * @param {object***REMOVED*** stream  A stream to write the serializiation to
 * @api public
 */
exports.writeToStream = function(object, options, stream) {
  if (typeof stream === 'undefined') {
    stream = options;
    options = {***REMOVED***;
  ***REMOVED***

  options = applyDefaults(object, options);

  return typeHasher(options, stream).dispatch(object);
***REMOVED***;

function typeHasher(options, writeTo, context){
  context = context || [];
  var write = function(str) {
    if (writeTo.update) {
      return writeTo.update(str, 'utf8');
    ***REMOVED*** else {
      return writeTo.write(str, 'utf8');
    ***REMOVED***
  ***REMOVED***;

  return {
    dispatch: function(value){
      if (options.replacer) {
        value = options.replacer(value);
      ***REMOVED***

      var type = typeof value;
      if (value === null) {
        type = 'null';
      ***REMOVED***

      //console.log("[DEBUG] Dispatch: ", value, "->", type, " -> ", "_" + type);

      return this['_' + type](value);
    ***REMOVED***,
    _object: function(object) {
      var pattern = (/\[object (.*)\]/i);
      var objString = Object.prototype.toString.call(object);
      var objType = pattern.exec(objString);
      if (!objType) { // object type did not match [object ...]
        objType = 'unknown:[' + objString + ']';
      ***REMOVED*** else {
        objType = objType[1]; // take only the class name
      ***REMOVED***

      objType = objType.toLowerCase();

      var objectNumber = null;

      if ((objectNumber = context.indexOf(object)) >= 0) {
        return this.dispatch('[CIRCULAR:' + objectNumber + ']');
      ***REMOVED*** else {
        context.push(object);
      ***REMOVED***

      if (typeof Buffer !== 'undefined' && Buffer.isBuffer && Buffer.isBuffer(object)) {
        write('buffer:');
        return write(object);
      ***REMOVED***

      if(objType !== 'object' && objType !== 'function' && objType !== 'asyncfunction') {
        if(this['_' + objType]) {
          this['_' + objType](object);
        ***REMOVED*** else if (options.ignoreUnknown) {
          return write('[' + objType + ']');
        ***REMOVED*** else {
          throw new Error('Unknown object type "' + objType + '"');
        ***REMOVED***
      ***REMOVED***else{
        var keys = Object.keys(object);
        if (options.unorderedObjects) {
          keys = keys.sort();
        ***REMOVED***
        // Make sure to incorporate special properties, so
        // Types with different prototypes will produce
        // a different hash and objects derived from
        // different functions (`new Foo`, `new Bar`) will
        // produce different hashes.
        // We never do this for native functions since some
        // seem to break because of that.
        if (options.respectType !== false && !isNativeFunction(object)) {
          keys.splice(0, 0, 'prototype', '__proto__', 'constructor');
        ***REMOVED***

        if (options.excludeKeys) {
          keys = keys.filter(function(key) { return !options.excludeKeys(key); ***REMOVED***);
        ***REMOVED***

        write('object:' + keys.length + ':');
        var self = this;
        return keys.forEach(function(key){
          self.dispatch(key);
          write(':');
          if(!options.excludeValues) {
            self.dispatch(object[key]);
          ***REMOVED***
          write(',');
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***,
    _array: function(arr, unordered){
      unordered = typeof unordered !== 'undefined' ? unordered :
        options.unorderedArrays !== false; // default to options.unorderedArrays

      var self = this;
      write('array:' + arr.length + ':');
      if (!unordered || arr.length <= 1) {
        return arr.forEach(function(entry) {
          return self.dispatch(entry);
        ***REMOVED***);
      ***REMOVED***

      // the unordered case is a little more complicated:
      // since there is no canonical ordering on objects,
      // i.e. {a:1***REMOVED*** < {a:2***REMOVED*** and {a:1***REMOVED*** > {a:2***REMOVED*** are both false,
      // we first serialize each entry using a PassThrough stream
      // before sorting.
      // also: we can’t use the same context array for all entries
      // since the order of hashing should *not* matter. instead,
      // we keep track of the additions to a copy of the context array
      // and add all of them to the global context array when we’re done
      var contextAdditions = [];
      var entries = arr.map(function(entry) {
        var strm = new PassThrough();
        var localContext = context.slice(); // make copy
        var hasher = typeHasher(options, strm, localContext);
        hasher.dispatch(entry);
        // take only what was added to localContext and append it to contextAdditions
        contextAdditions = contextAdditions.concat(localContext.slice(context.length));
        return strm.read().toString();
      ***REMOVED***);
      context = context.concat(contextAdditions);
      entries.sort();
      return this._array(entries, false);
    ***REMOVED***,
    _date: function(date){
      return write('date:' + date.toJSON());
    ***REMOVED***,
    _symbol: function(sym){
      return write('symbol:' + sym.toString());
    ***REMOVED***,
    _error: function(err){
      return write('error:' + err.toString());
    ***REMOVED***,
    _boolean: function(bool){
      return write('bool:' + bool.toString());
    ***REMOVED***,
    _string: function(string){
      write('string:' + string.length + ':');
      write(string.toString());
    ***REMOVED***,
    _function: function(fn){
      write('fn:');
      if (isNativeFunction(fn)) {
        this.dispatch('[native]');
      ***REMOVED*** else {
        this.dispatch(fn.toString());
      ***REMOVED***

      if (options.respectFunctionNames !== false) {
        // Make sure we can still distinguish native functions
        // by their name, otherwise String and Function will
        // have the same hash
        this.dispatch("function-name:" + String(fn.name));
      ***REMOVED***

      if (options.respectFunctionProperties) {
        this._object(fn);
      ***REMOVED***
    ***REMOVED***,
    _number: function(number){
      return write('number:' + number.toString());
    ***REMOVED***,
    _xml: function(xml){
      return write('xml:' + xml.toString());
    ***REMOVED***,
    _null: function() {
      return write('Null');
    ***REMOVED***,
    _undefined: function() {
      return write('Undefined');
    ***REMOVED***,
    _regexp: function(regex){
      return write('regex:' + regex.toString());
    ***REMOVED***,
    _uint8array: function(arr){
      write('uint8array:');
      return this.dispatch(Array.prototype.slice.call(arr));
    ***REMOVED***,
    _uint8clampedarray: function(arr){
      write('uint8clampedarray:');
      return this.dispatch(Array.prototype.slice.call(arr));
    ***REMOVED***,
    _int8array: function(arr){
      write('int8array:');
      return this.dispatch(Array.prototype.slice.call(arr));
    ***REMOVED***,
    _uint16array: function(arr){
      write('uint16array:');
      return this.dispatch(Array.prototype.slice.call(arr));
    ***REMOVED***,
    _int16array: function(arr){
      write('int16array:');
      return this.dispatch(Array.prototype.slice.call(arr));
    ***REMOVED***,
    _uint32array: function(arr){
      write('uint32array:');
      return this.dispatch(Array.prototype.slice.call(arr));
    ***REMOVED***,
    _int32array: function(arr){
      write('int32array:');
      return this.dispatch(Array.prototype.slice.call(arr));
    ***REMOVED***,
    _float32array: function(arr){
      write('float32array:');
      return this.dispatch(Array.prototype.slice.call(arr));
    ***REMOVED***,
    _float64array: function(arr){
      write('float64array:');
      return this.dispatch(Array.prototype.slice.call(arr));
    ***REMOVED***,
    _arraybuffer: function(arr){
      write('arraybuffer:');
      return this.dispatch(new Uint8Array(arr));
    ***REMOVED***,
    _url: function(url) {
      return write('url:' + url.toString(), 'utf8');
    ***REMOVED***,
    _map: function(map) {
      write('map:');
      var arr = Array.from(map);
      return this._array(arr, options.unorderedSets !== false);
    ***REMOVED***,
    _set: function(set) {
      write('set:');
      var arr = Array.from(set);
      return this._array(arr, options.unorderedSets !== false);
    ***REMOVED***,
    _file: function(file) {
      write('file:');
      return this.dispatch([file.name, file.size, file.type, file.lastModfied]);
    ***REMOVED***,
    _blob: function() {
      if (options.ignoreUnknown) {
        return write('[blob]');
      ***REMOVED***

      throw Error('Hashing Blob objects is currently not supported\n' +
        '(see https://github.com/puleos/object-hash/issues/26)\n' +
        'Use "options.replacer" or "options.ignoreUnknown"\n');
    ***REMOVED***,
    _domwindow: function() { return write('domwindow'); ***REMOVED***,
    _bigint: function(number){
      return write('bigint:' + number.toString());
    ***REMOVED***,
    /* Node.js standard native objects */
    _process: function() { return write('process'); ***REMOVED***,
    _timer: function() { return write('timer'); ***REMOVED***,
    _pipe: function() { return write('pipe'); ***REMOVED***,
    _tcp: function() { return write('tcp'); ***REMOVED***,
    _udp: function() { return write('udp'); ***REMOVED***,
    _tty: function() { return write('tty'); ***REMOVED***,
    _statwatcher: function() { return write('statwatcher'); ***REMOVED***,
    _securecontext: function() { return write('securecontext'); ***REMOVED***,
    _connection: function() { return write('connection'); ***REMOVED***,
    _zlib: function() { return write('zlib'); ***REMOVED***,
    _context: function() { return write('context'); ***REMOVED***,
    _nodescript: function() { return write('nodescript'); ***REMOVED***,
    _httpparser: function() { return write('httpparser'); ***REMOVED***,
    _dataview: function() { return write('dataview'); ***REMOVED***,
    _signal: function() { return write('signal'); ***REMOVED***,
    _fsevent: function() { return write('fsevent'); ***REMOVED***,
    _tlswrap: function() { return write('tlswrap'); ***REMOVED***,
  ***REMOVED***;
***REMOVED***

// Mini-implementation of stream.PassThrough
// We are far from having need for the full implementation, and we can
// make assumptions like "many writes, then only one final read"
// and we can ignore encoding specifics
function PassThrough() {
  return {
    buf: '',

    write: function(b) {
      this.buf += b;
    ***REMOVED***,

    end: function(b) {
      this.buf += b;
    ***REMOVED***,

    read: function() {
      return this.buf;
    ***REMOVED***
  ***REMOVED***;
***REMOVED***
