// @babel/traverse/lib/path/family.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports._getKey = _getKey;
exports._getPattern = _getPattern;
exports.get = get;
exports.getAllNextSiblings = getAllNextSiblings;
exports.getAllPrevSiblings = getAllPrevSiblings;
exports.getAssignmentIdentifiers = getAssignmentIdentifiers;
exports.getBindingIdentifierPaths = getBindingIdentifierPaths;
exports.getBindingIdentifiers = getBindingIdentifiers;
exports.getCompletionRecords = getCompletionRecords;
exports.getNextSibling = getNextSibling;
exports.getOpposite = getOpposite;
exports.getOuterBindingIdentifierPaths = getOuterBindingIdentifierPaths;
exports.getOuterBindingIdentifiers = getOuterBindingIdentifiers;
exports.getPrevSibling = getPrevSibling;
exports.getSibling = getSibling;
var _index = require("./index.js");
var _t = require("@babel/types");
const {
  getAssignmentIdentifiers: _getAssignmentIdentifiers,
  getBindingIdentifiers: _getBindingIdentifiers,
  getOuterBindingIdentifiers: _getOuterBindingIdentifiers,
  numericLiteral,
  unaryExpression
***REMOVED*** = _t;
const NORMAL_COMPLETION = 0;
const BREAK_COMPLETION = 1;
function NormalCompletion(path) {
  return {
    type: NORMAL_COMPLETION,
    path
  ***REMOVED***;
***REMOVED***
function BreakCompletion(path) {
  return {
    type: BREAK_COMPLETION,
    path
  ***REMOVED***;
***REMOVED***
function getOpposite() {
  if (this.key === "left") {
    return this.getSibling("right");
  ***REMOVED*** else if (this.key === "right") {
    return this.getSibling("left");
  ***REMOVED***
  return null;
***REMOVED***
function addCompletionRecords(path, records, context) {
  if (path) {
    records.push(..._getCompletionRecords(path, context));
  ***REMOVED***
  return records;
***REMOVED***
function completionRecordForSwitch(cases, records, context) {
  let lastNormalCompletions = [];
  for (let i = 0; i < cases.length; i++) {
    const casePath = cases[i];
    const caseCompletions = _getCompletionRecords(casePath, context);
    const normalCompletions = [];
    const breakCompletions = [];
    for (const c of caseCompletions) {
      if (c.type === NORMAL_COMPLETION) {
        normalCompletions.push(c);
      ***REMOVED***
      if (c.type === BREAK_COMPLETION) {
        breakCompletions.push(c);
      ***REMOVED***
    ***REMOVED***
    if (normalCompletions.length) {
      lastNormalCompletions = normalCompletions;
    ***REMOVED***
    records.push(...breakCompletions);
  ***REMOVED***
  records.push(...lastNormalCompletions);
  return records;
***REMOVED***
function normalCompletionToBreak(completions) {
  completions.forEach(c => {
    c.type = BREAK_COMPLETION;
  ***REMOVED***);
***REMOVED***
function replaceBreakStatementInBreakCompletion(completions, reachable) {
  completions.forEach(c => {
    if (c.path.isBreakStatement({
      label: null
    ***REMOVED***)) {
      if (reachable) {
        c.path.replaceWith(unaryExpression("void", numericLiteral(0)));
      ***REMOVED*** else {
        c.path.remove();
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***);
***REMOVED***
function getStatementListCompletion(paths, context) {
  const completions = [];
  if (context.canHaveBreak) {
    let lastNormalCompletions = [];
    for (let i = 0; i < paths.length; i++) {
      const path = paths[i];
      const newContext = Object.assign({***REMOVED***, context, {
        inCaseClause: false
      ***REMOVED***);
      if (path.isBlockStatement() && (context.inCaseClause || context.shouldPopulateBreak)) {
        newContext.shouldPopulateBreak = true;
      ***REMOVED*** else {
        newContext.shouldPopulateBreak = false;
      ***REMOVED***
      const statementCompletions = _getCompletionRecords(path, newContext);
      if (statementCompletions.length > 0 && statementCompletions.every(c => c.type === BREAK_COMPLETION)) {
        if (lastNormalCompletions.length > 0 && statementCompletions.every(c => c.path.isBreakStatement({
          label: null
        ***REMOVED***))) {
          normalCompletionToBreak(lastNormalCompletions);
          completions.push(...lastNormalCompletions);
          if (lastNormalCompletions.some(c => c.path.isDeclaration())) {
            completions.push(...statementCompletions);
            if (!context.shouldPreserveBreak) {
              replaceBreakStatementInBreakCompletion(statementCompletions, true);
            ***REMOVED***
          ***REMOVED***
          if (!context.shouldPreserveBreak) {
            replaceBreakStatementInBreakCompletion(statementCompletions, false);
          ***REMOVED***
        ***REMOVED*** else {
          completions.push(...statementCompletions);
          if (!context.shouldPopulateBreak && !context.shouldPreserveBreak) {
            replaceBreakStatementInBreakCompletion(statementCompletions, true);
          ***REMOVED***
        ***REMOVED***
        break;
      ***REMOVED***
      if (i === paths.length - 1) {
        completions.push(...statementCompletions);
      ***REMOVED*** else {
        lastNormalCompletions = [];
        for (let i = 0; i < statementCompletions.length; i++) {
          const c = statementCompletions[i];
          if (c.type === BREAK_COMPLETION) {
            completions.push(c);
          ***REMOVED***
          if (c.type === NORMAL_COMPLETION) {
            lastNormalCompletions.push(c);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED*** else if (paths.length) {
    for (let i = paths.length - 1; i >= 0; i--) {
      const pathCompletions = _getCompletionRecords(paths[i], context);
      if (pathCompletions.length > 1 || pathCompletions.length === 1 && !pathCompletions[0].path.isVariableDeclaration() && !pathCompletions[0].path.isEmptyStatement()) {
        completions.push(...pathCompletions);
        break;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  return completions;
***REMOVED***
function _getCompletionRecords(path, context) {
  let records = [];
  if (path.isIfStatement()) {
    records = addCompletionRecords(path.get("consequent"), records, context);
    records = addCompletionRecords(path.get("alternate"), records, context);
  ***REMOVED*** else if (path.isDoExpression() || path.isFor() || path.isWhile() || path.isLabeledStatement()) {
    return addCompletionRecords(path.get("body"), records, context);
  ***REMOVED*** else if (path.isProgram() || path.isBlockStatement()) {
    return getStatementListCompletion(path.get("body"), context);
  ***REMOVED*** else if (path.isFunction()) {
    return _getCompletionRecords(path.get("body"), context);
  ***REMOVED*** else if (path.isTryStatement()) {
    records = addCompletionRecords(path.get("block"), records, context);
    records = addCompletionRecords(path.get("handler"), records, context);
  ***REMOVED*** else if (path.isCatchClause()) {
    return addCompletionRecords(path.get("body"), records, context);
  ***REMOVED*** else if (path.isSwitchStatement()) {
    return completionRecordForSwitch(path.get("cases"), records, context);
  ***REMOVED*** else if (path.isSwitchCase()) {
    return getStatementListCompletion(path.get("consequent"), {
      canHaveBreak: true,
      shouldPopulateBreak: false,
      inCaseClause: true,
      shouldPreserveBreak: context.shouldPreserveBreak
    ***REMOVED***);
  ***REMOVED*** else if (path.isBreakStatement()) {
    records.push(BreakCompletion(path));
  ***REMOVED*** else {
    records.push(NormalCompletion(path));
  ***REMOVED***
  return records;
***REMOVED***
function getCompletionRecords(shouldPreserveBreak = false) {
  const records = _getCompletionRecords(this, {
    canHaveBreak: false,
    shouldPopulateBreak: false,
    inCaseClause: false,
    shouldPreserveBreak
  ***REMOVED***);
  return records.map(r => r.path);
***REMOVED***
function getSibling(key) {
  return _index.default.get({
    parentPath: this.parentPath,
    parent: this.parent,
    container: this.container,
    listKey: this.listKey,
    key: key
  ***REMOVED***).setContext(this.context);
***REMOVED***
function getPrevSibling() {
  return this.getSibling(this.key - 1);
***REMOVED***
function getNextSibling() {
  return this.getSibling(this.key + 1);
***REMOVED***
function getAllNextSiblings() {
  let _key = this.key;
  let sibling = this.getSibling(++_key);
  const siblings = [];
  while (sibling.node) {
    siblings.push(sibling);
    sibling = this.getSibling(++_key);
  ***REMOVED***
  return siblings;
***REMOVED***
function getAllPrevSiblings() {
  let _key = this.key;
  let sibling = this.getSibling(--_key);
  const siblings = [];
  while (sibling.node) {
    siblings.push(sibling);
    sibling = this.getSibling(--_key);
  ***REMOVED***
  return siblings;
***REMOVED***
function get(key, context = true) {
  if (context === true) context = this.context;
  const parts = key.split(".");
  if (parts.length === 1) {
    return _getKey.call(this, key, context);
  ***REMOVED*** else {
    return _getPattern.call(this, parts, context);
  ***REMOVED***
***REMOVED***
function _getKey(key, context) {
  const node = this.node;
  const container = node[key];
  if (Array.isArray(container)) {
    return container.map((_, i) => {
      return _index.default.get({
        listKey: key,
        parentPath: this,
        parent: node,
        container: container,
        key: i
      ***REMOVED***).setContext(context);
    ***REMOVED***);
  ***REMOVED*** else {
    return _index.default.get({
      parentPath: this,
      parent: node,
      container: node,
      key: key
    ***REMOVED***).setContext(context);
  ***REMOVED***
***REMOVED***
function _getPattern(parts, context) {
  let path = this;
  for (const part of parts) {
    if (part === ".") {
      path = path.parentPath;
    ***REMOVED*** else {
      if (Array.isArray(path)) {
        path = path[part];
      ***REMOVED*** else {
        path = path.get(part, context);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  return path;
***REMOVED***
function getAssignmentIdentifiers() {
  return _getAssignmentIdentifiers(this.node);
***REMOVED***
function getBindingIdentifiers(duplicates) {
  return _getBindingIdentifiers(this.node, duplicates);
***REMOVED***
function getOuterBindingIdentifiers(duplicates) {
  return _getOuterBindingIdentifiers(this.node, duplicates);
***REMOVED***
function getBindingIdentifierPaths(duplicates = false, outerOnly = false) {
  const path = this;
  const search = [path];
  const ids = Object.create(null);
  while (search.length) {
    const id = search.shift();
    if (!id) continue;
    if (!id.node) continue;
    const keys = _getBindingIdentifiers.keys[id.node.type];
    if (id.isIdentifier()) {
      if (duplicates) {
        const _ids = ids[id.node.name] = ids[id.node.name] || [];
        _ids.push(id);
      ***REMOVED*** else {
        ids[id.node.name] = id;
      ***REMOVED***
      continue;
    ***REMOVED***
    if (id.isExportDeclaration()) {
      const declaration = id.get("declaration");
      if (declaration.isDeclaration()) {
        search.push(declaration);
      ***REMOVED***
      continue;
    ***REMOVED***
    if (outerOnly) {
      if (id.isFunctionDeclaration()) {
        search.push(id.get("id"));
        continue;
      ***REMOVED***
      if (id.isFunctionExpression()) {
        continue;
      ***REMOVED***
    ***REMOVED***
    if (keys) {
      for (let i = 0; i < keys.length; i++) {
        const key = keys[i];
        const child = id.get(key);
        if (Array.isArray(child)) {
          search.push(...child);
        ***REMOVED*** else if (child.node) {
          search.push(child);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  return ids;
***REMOVED***
function getOuterBindingIdentifierPaths(duplicates = false) {
  return this.getBindingIdentifierPaths(duplicates, true);
***REMOVED***
