// postcss-nested/index.js
const { AtRule, Rule ***REMOVED*** = require('postcss')
let parser = require('postcss-selector-parser')

/**
 * Run a selector string through postcss-selector-parser
 */
function parse(rawSelector, rule) {
  let nodes
  try {
    parser(parsed => {
      nodes = parsed
    ***REMOVED***).processSync(rawSelector)
  ***REMOVED*** catch (e) {
    if (rawSelector.includes(':')) {
      throw rule ? rule.error('Missed semicolon') : e
    ***REMOVED*** else {
      throw rule ? rule.error(e.message) : e
    ***REMOVED***
  ***REMOVED***
  return nodes.at(0)
***REMOVED***

/**
 * Replaces the "&" token in a node's selector with the parent selector
 * similar to what SCSS does.
 *
 * Mutates the nodes list
 */
function interpolateAmpInSelector(nodes, parent) {
  let replaced = false
  nodes.each(node => {
    if (node.type === 'nesting') {
      let clonedParent = parent.clone({***REMOVED***)
      if (node.value !== '&') {
        node.replaceWith(
          parse(node.value.replace('&', clonedParent.toString()))
        )
      ***REMOVED*** else {
        node.replaceWith(clonedParent)
      ***REMOVED***
      replaced = true
    ***REMOVED*** else if ('nodes' in node && node.nodes) {
      if (interpolateAmpInSelector(node, parent)) {
        replaced = true
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***)
  return replaced
***REMOVED***

/**
 * Combines parent and child selectors, in a SCSS-like way
 */
function mergeSelectors(parent, child) {
  let merged = []
  parent.selectors.forEach(sel => {
    let parentNode = parse(sel, parent)

    child.selectors.forEach(selector => {
      if (!selector) {
        return
      ***REMOVED***
      let node = parse(selector, child)
      let replaced = interpolateAmpInSelector(node, parentNode)
      if (!replaced) {
        node.prepend(parser.combinator({ value: ' ' ***REMOVED***))
        node.prepend(parentNode.clone({***REMOVED***))
      ***REMOVED***
      merged.push(node.toString())
    ***REMOVED***)
  ***REMOVED***)
  return merged
***REMOVED***

/**
 * Move a child and its preceeding comment(s) to after "after"
 */
function breakOut(child, after) {
  let prev = child.prev()
  after.after(child)
  while (prev && prev.type === 'comment') {
    let nextPrev = prev.prev()
    after.after(prev)
    prev = nextPrev
  ***REMOVED***
  return child
***REMOVED***

function createFnAtruleChilds(bubble) {
  return function atruleChilds(rule, atrule, bubbling, mergeSels = bubbling) {
    let children = []
    atrule.each(child => {
      if (child.type === 'rule' && bubbling) {
        if (mergeSels) {
          child.selectors = mergeSelectors(rule, child)
        ***REMOVED***
      ***REMOVED*** else if (child.type === 'atrule' && child.nodes) {
        if (bubble[child.name]) {
          atruleChilds(rule, child, mergeSels)
        ***REMOVED*** else if (atrule[rootRuleMergeSel] !== false) {
          children.push(child)
        ***REMOVED***
      ***REMOVED*** else {
        children.push(child)
      ***REMOVED***
    ***REMOVED***)
    if (bubbling) {
      if (children.length) {
        let clone = rule.clone({ nodes: [] ***REMOVED***)
        for (let child of children) {
          clone.append(child)
        ***REMOVED***
        atrule.prepend(clone)
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***

function pickDeclarations(selector, declarations, after) {
  let parent = new Rule({
    nodes: [],
    selector
  ***REMOVED***)
  parent.append(declarations)
  after.after(parent)
  return parent
***REMOVED***

function atruleNames(defaults, custom) {
  let list = {***REMOVED***
  for (let name of defaults) {
    list[name] = true
  ***REMOVED***
  if (custom) {
    for (let name of custom) {
      list[name.replace(/^@/, '')] = true
    ***REMOVED***
  ***REMOVED***
  return list
***REMOVED***

function parseRootRuleParams(params) {
  params = params.trim()
  let braceBlock = params.match(/^\((.*)\)$/)
  if (!braceBlock) {
    return { selector: params, type: 'basic' ***REMOVED***
  ***REMOVED***
  let bits = braceBlock[1].match(/^(with(?:out)?):(.+)$/)
  if (bits) {
    let allowlist = bits[1] === 'with'
    let rules = Object.fromEntries(
      bits[2]
        .trim()
        .split(/\s+/)
        .map(name => [name, true])
    )
    if (allowlist && rules.all) {
      return { type: 'noop' ***REMOVED***
    ***REMOVED***
    let escapes = rule => !!rules[rule]
    if (rules.all) {
      escapes = () => true
    ***REMOVED*** else if (allowlist) {
      escapes = rule => (rule === 'all' ? false : !rules[rule])
    ***REMOVED***

    return {
      escapes,
      type: 'withrules'
    ***REMOVED***
  ***REMOVED***
  // Unrecognized brace block
  return { type: 'unknown' ***REMOVED***
***REMOVED***

function getAncestorRules(leaf) {
  let lineage = []
  let parent = leaf.parent

  while (parent && parent instanceof AtRule) {
    lineage.push(parent)
    parent = parent.parent
  ***REMOVED***
  return lineage
***REMOVED***

function unwrapRootRule(rule) {
  let escapes = rule[rootRuleEscapes]

  if (!escapes) {
    rule.after(rule.nodes)
  ***REMOVED*** else {
    let nodes = rule.nodes

    let topEscaped
    let topEscapedIdx = -1
    let breakoutLeaf
    let breakoutRoot
    let clone

    let lineage = getAncestorRules(rule)
    lineage.forEach((parent, i) => {
      if (escapes(parent.name)) {
        topEscaped = parent
        topEscapedIdx = i
        breakoutRoot = clone
      ***REMOVED*** else {
        let oldClone = clone
        clone = parent.clone({ nodes: [] ***REMOVED***)
        oldClone && clone.append(oldClone)
        breakoutLeaf = breakoutLeaf || clone
      ***REMOVED***
    ***REMOVED***)

    if (!topEscaped) {
      rule.after(nodes)
    ***REMOVED*** else if (!breakoutRoot) {
      topEscaped.after(nodes)
    ***REMOVED*** else {
      let leaf = breakoutLeaf
      leaf.append(nodes)
      topEscaped.after(breakoutRoot)
    ***REMOVED***

    if (rule.next() && topEscaped) {
      let restRoot
      lineage.slice(0, topEscapedIdx + 1).forEach((parent, i, arr) => {
        let oldRoot = restRoot
        restRoot = parent.clone({ nodes: [] ***REMOVED***)
        oldRoot && restRoot.append(oldRoot)

        let nextSibs = []
        let _child = arr[i - 1] || rule
        let next = _child.next()
        while (next) {
          nextSibs.push(next)
          next = next.next()
        ***REMOVED***
        restRoot.append(nextSibs)
      ***REMOVED***)
      restRoot && (breakoutRoot || nodes[nodes.length - 1]).after(restRoot)
    ***REMOVED***
  ***REMOVED***

  rule.remove()
***REMOVED***

const rootRuleMergeSel = Symbol('rootRuleMergeSel')
const rootRuleEscapes = Symbol('rootRuleEscapes')

function normalizeRootRule(rule) {
  let { params ***REMOVED*** = rule
  let { escapes, selector, type ***REMOVED*** = parseRootRuleParams(params)
  if (type === 'unknown') {
    throw rule.error(
      `Unknown @${rule.name***REMOVED*** parameter ${JSON.stringify(params)***REMOVED***`
    )
  ***REMOVED***
  if (type === 'basic' && selector) {
    let selectorBlock = new Rule({ nodes: rule.nodes, selector ***REMOVED***)
    rule.removeAll()
    rule.append(selectorBlock)
  ***REMOVED***
  rule[rootRuleEscapes] = escapes
  rule[rootRuleMergeSel] = escapes ? !escapes('all') : type === 'noop'
***REMOVED***

const hasRootRule = Symbol('hasRootRule')

module.exports = (opts = {***REMOVED***) => {
  let bubble = atruleNames(
    ['media', 'supports', 'layer', 'container', 'starting-style'],
    opts.bubble
  )
  let atruleChilds = createFnAtruleChilds(bubble)
  let unwrap = atruleNames(
    [
      'document',
      'font-face',
      'keyframes',
      '-webkit-keyframes',
      '-moz-keyframes'
    ],
    opts.unwrap
  )
  let rootRuleName = (opts.rootRuleName || 'at-root').replace(/^@/, '')
  let preserveEmpty = opts.preserveEmpty

  return {
    Once(root) {
      root.walkAtRules(rootRuleName, node => {
        normalizeRootRule(node)
        root[hasRootRule] = true
      ***REMOVED***)
    ***REMOVED***,

    postcssPlugin: 'postcss-nested',

    RootExit(root) {
      if (root[hasRootRule]) {
        root.walkAtRules(rootRuleName, unwrapRootRule)
        root[hasRootRule] = false
      ***REMOVED***
    ***REMOVED***,

    Rule(rule) {
      let unwrapped = false
      let after = rule
      let copyDeclarations = false
      let declarations = []

      rule.each(child => {
        if (child.type === 'rule') {
          if (declarations.length) {
            after = pickDeclarations(rule.selector, declarations, after)
            declarations = []
          ***REMOVED***

          copyDeclarations = true
          unwrapped = true
          child.selectors = mergeSelectors(rule, child)
          after = breakOut(child, after)
        ***REMOVED*** else if (child.type === 'atrule') {
          if (declarations.length) {
            after = pickDeclarations(rule.selector, declarations, after)
            declarations = []
          ***REMOVED***
          if (child.name === rootRuleName) {
            unwrapped = true
            atruleChilds(rule, child, true, child[rootRuleMergeSel])
            after = breakOut(child, after)
          ***REMOVED*** else if (bubble[child.name]) {
            copyDeclarations = true
            unwrapped = true
            atruleChilds(rule, child, true)
            after = breakOut(child, after)
          ***REMOVED*** else if (unwrap[child.name]) {
            copyDeclarations = true
            unwrapped = true
            atruleChilds(rule, child, false)
            after = breakOut(child, after)
          ***REMOVED*** else if (copyDeclarations) {
            declarations.push(child)
          ***REMOVED***
        ***REMOVED*** else if (child.type === 'decl' && copyDeclarations) {
          declarations.push(child)
        ***REMOVED***
      ***REMOVED***)

      if (declarations.length) {
        after = pickDeclarations(rule.selector, declarations, after)
      ***REMOVED***

      if (unwrapped && preserveEmpty !== true) {
        rule.raws.semicolon = true
        if (rule.nodes.length === 0) rule.remove()
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***
module.exports.postcss = true
