// aria-query/lib/etc/roles/graphics/graphicsDocumentRole.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var graphicsDocumentRole = {
  abstract: false,
  accessibleNameRequired: true,
  baseConcepts: [],
  childrenPresentational: false,
  nameFrom: ['author'],
  prohibitedProps: [],
  props: {
    'aria-disabled': null,
    'aria-errormessage': null,
    'aria-expanded': null,
    'aria-haspopup': null,
    'aria-invalid': null
  ***REMOVED***,
  relatedConcepts: [{
    module: 'GRAPHICS',
    concept: {
      name: 'graphics-object'
    ***REMOVED***
  ***REMOVED***, {
    module: 'ARIA',
    concept: {
      name: 'img'
    ***REMOVED***
  ***REMOVED***, {
    module: 'ARIA',
    concept: {
      name: 'article'
    ***REMOVED***
  ***REMOVED***],
  requireContextRole: [],
  requiredContextRole: [],
  requiredOwnedElements: [],
  requiredProps: {***REMOVED***,
  superClass: [['roletype', 'structure', 'document']]
***REMOVED***;
var _default = exports.default = graphicsDocumentRole;
