// @babel/plugin-transform-parameters/lib/rest.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = convertFunctionRest;
var _core = require("@babel/core");
var _shadowUtils = require("./shadow-utils.js");
const buildRest = _core.template.statement(`
  for (var LEN = ARGUMENTS.length,
           ARRAY = new Array(ARRAY_LEN),
           KEY = START;
       KEY < LEN;
       KEY++) {
    ARRAY[ARRAY_KEY] = ARGUMENTS[KEY];
  ***REMOVED***
`);
const restIndex = _core.template.expression(`
  (INDEX < OFFSET || ARGUMENTS.length <= INDEX) ? undefined : ARGUMENTS[INDEX]
`);
const restIndexImpure = _core.template.expression(`
  REF = INDEX, (REF < OFFSET || ARGUMENTS.length <= REF) ? undefined : ARGUMENTS[REF]
`);
const restLength = _core.template.expression(`
  ARGUMENTS.length <= OFFSET ? 0 : ARGUMENTS.length - OFFSET
`);
function referencesRest(path, state) {
  if (path.node.name === state.name) {
    return path.scope.bindingIdentifierEquals(state.name, state.outerBinding);
  ***REMOVED***
  return false;
***REMOVED***
const memberExpressionOptimisationVisitor = {
  Scope(path, state) {
    if (!path.scope.bindingIdentifierEquals(state.name, state.outerBinding)) {
      path.skip();
    ***REMOVED***
  ***REMOVED***,
  Flow(path) {
    if (path.isTypeCastExpression()) return;
    path.skip();
  ***REMOVED***,
  Function(path, state) {
    const oldNoOptimise = state.noOptimise;
    state.noOptimise = true;
    path.traverse(memberExpressionOptimisationVisitor, state);
    state.noOptimise = oldNoOptimise;
    path.skip();
  ***REMOVED***,
  ReferencedIdentifier(path, state) {
    const {
      node
    ***REMOVED*** = path;
    if (node.name === "arguments") {
      state.deopted = true;
    ***REMOVED***
    if (!referencesRest(path, state)) return;
    if (state.noOptimise) {
      state.deopted = true;
    ***REMOVED*** else {
      const {
        parentPath
      ***REMOVED*** = path;
      if (parentPath.listKey === "params" && parentPath.key < state.offset) {
        return;
      ***REMOVED***
      if (parentPath.isMemberExpression({
        object: node
      ***REMOVED***)) {
        const grandparentPath = parentPath.parentPath;
        const argsOptEligible = !state.deopted && !(grandparentPath.isAssignmentExpression() && parentPath.node === grandparentPath.node.left || grandparentPath.isLVal() || grandparentPath.isForXStatement() || grandparentPath.isUpdateExpression() || grandparentPath.isUnaryExpression({
          operator: "delete"
        ***REMOVED***) || (grandparentPath.isCallExpression() || grandparentPath.isNewExpression()) && parentPath.node === grandparentPath.node.callee);
        if (argsOptEligible) {
          if (parentPath.node.computed) {
            if (parentPath.get("property").isBaseType("number")) {
              state.candidates.push({
                cause: "indexGetter",
                path
              ***REMOVED***);
              return;
            ***REMOVED***
          ***REMOVED*** else if (parentPath.node.property.name === "length") {
            state.candidates.push({
              cause: "lengthGetter",
              path
            ***REMOVED***);
            return;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
      if (state.offset === 0 && parentPath.isSpreadElement()) {
        const call = parentPath.parentPath;
        if (call.isCallExpression() && call.node.arguments.length === 1) {
          state.candidates.push({
            cause: "argSpread",
            path
          ***REMOVED***);
          return;
        ***REMOVED***
      ***REMOVED***
      state.references.push(path);
    ***REMOVED***
  ***REMOVED***,
  BindingIdentifier(path, state) {
    if (referencesRest(path, state)) {
      state.deopted = true;
    ***REMOVED***
  ***REMOVED***
***REMOVED***;
function getParamsCount(node) {
  let count = node.params.length;
  if (count > 0 && _core.types.isIdentifier(node.params[0], {
    name: "this"
  ***REMOVED***)) {
    count -= 1;
  ***REMOVED***
  return count;
***REMOVED***
function hasRest(node) {
  const length = node.params.length;
  return length > 0 && _core.types.isRestElement(node.params[length - 1]);
***REMOVED***
function optimiseIndexGetter(path, argsId, offset) {
  const offsetLiteral = _core.types.numericLiteral(offset);
  let index;
  const parent = path.parent;
  if (_core.types.isNumericLiteral(parent.property)) {
    index = _core.types.numericLiteral(parent.property.value + offset);
  ***REMOVED*** else if (offset === 0) {
    index = parent.property;
  ***REMOVED*** else {
    index = _core.types.binaryExpression("+", parent.property, _core.types.cloneNode(offsetLiteral));
  ***REMOVED***
  const {
    scope,
    parentPath
  ***REMOVED*** = path;
  if (!scope.isPure(index)) {
    const temp = scope.generateUidIdentifierBasedOnNode(index);
    scope.push({
      id: temp,
      kind: "var"
    ***REMOVED***);
    parentPath.replaceWith(restIndexImpure({
      ARGUMENTS: argsId,
      OFFSET: offsetLiteral,
      INDEX: index,
      REF: _core.types.cloneNode(temp)
    ***REMOVED***));
  ***REMOVED*** else {
    parentPath.replaceWith(restIndex({
      ARGUMENTS: argsId,
      OFFSET: offsetLiteral,
      INDEX: index
    ***REMOVED***));
    const replacedParentPath = parentPath;
    const offsetTestPath = replacedParentPath.get("test");
    const valRes = offsetTestPath.get("left").evaluate();
    if (valRes.confident) {
      if (valRes.value === true) {
        replacedParentPath.replaceWith(scope.buildUndefinedNode());
      ***REMOVED*** else {
        offsetTestPath.replaceWith(offsetTestPath.get("right"));
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***
function optimiseLengthGetter(path, argsId, offset) {
  if (offset) {
    path.parentPath.replaceWith(restLength({
      ARGUMENTS: argsId,
      OFFSET: _core.types.numericLiteral(offset)
    ***REMOVED***));
  ***REMOVED*** else {
    path.replaceWith(argsId);
  ***REMOVED***
***REMOVED***
function convertFunctionRest(path) {
  const {
    node,
    scope
  ***REMOVED*** = path;
  if (!hasRest(node)) return false;
  const restPath = path.get(`params.${node.params.length - 1***REMOVED***.argument`);
  if (!restPath.isIdentifier()) {
    const shadowedParams = new Set();
    (0, _shadowUtils.collectShadowedParamsNames)(restPath, path.scope, shadowedParams);
    let needsIIFE = shadowedParams.size > 0;
    if (!needsIIFE) {
      const state = {
        needsOuterBinding: false,
        scope
      ***REMOVED***;
      restPath.traverse(_shadowUtils.iifeVisitor, state);
      needsIIFE = state.needsOuterBinding;
    ***REMOVED***
    if (needsIIFE) {
      path.ensureBlock();
      path.set("body", _core.types.blockStatement([(0, _shadowUtils.buildScopeIIFE)(shadowedParams, path.node.body)]));
    ***REMOVED***
  ***REMOVED***
  let rest = restPath.node;
  node.params.pop();
  if (_core.types.isPattern(rest)) {
    const pattern = rest;
    rest = scope.generateUidIdentifier("ref");
    const declar = _core.types.variableDeclaration("let", [_core.types.variableDeclarator(pattern, rest)]);
    path.ensureBlock();
    node.body.body.unshift(declar);
  ***REMOVED*** else if (rest.name === "arguments") {
    scope.rename(rest.name);
  ***REMOVED***
  const argsId = _core.types.identifier("arguments");
  const paramsCount = getParamsCount(node);
  const state = {
    references: [],
    offset: paramsCount,
    argumentsNode: argsId,
    outerBinding: scope.getBindingIdentifier(rest.name),
    candidates: [],
    name: rest.name,
    deopted: false
  ***REMOVED***;
  path.traverse(memberExpressionOptimisationVisitor, state);
  if (!state.deopted && !state.references.length) {
    for (const {
      path,
      cause
    ***REMOVED*** of state.candidates) {
      const clonedArgsId = _core.types.cloneNode(argsId);
      switch (cause) {
        case "indexGetter":
          optimiseIndexGetter(path, clonedArgsId, state.offset);
          break;
        case "lengthGetter":
          optimiseLengthGetter(path, clonedArgsId, state.offset);
          break;
        default:
          path.replaceWith(clonedArgsId);
      ***REMOVED***
    ***REMOVED***
    return true;
  ***REMOVED***
  state.references.push(...state.candidates.map(({
    path
  ***REMOVED***) => path));
  const start = _core.types.numericLiteral(paramsCount);
  const key = scope.generateUidIdentifier("key");
  const len = scope.generateUidIdentifier("len");
  let arrKey, arrLen;
  if (paramsCount) {
    arrKey = _core.types.binaryExpression("-", _core.types.cloneNode(key), _core.types.cloneNode(start));
    arrLen = _core.types.conditionalExpression(_core.types.binaryExpression(">", _core.types.cloneNode(len), _core.types.cloneNode(start)), _core.types.binaryExpression("-", _core.types.cloneNode(len), _core.types.cloneNode(start)), _core.types.numericLiteral(0));
  ***REMOVED*** else {
    arrKey = _core.types.identifier(key.name);
    arrLen = _core.types.identifier(len.name);
  ***REMOVED***
  const loop = buildRest({
    ARGUMENTS: argsId,
    ARRAY_KEY: arrKey,
    ARRAY_LEN: arrLen,
    START: start,
    ARRAY: rest,
    KEY: key,
    LEN: len
  ***REMOVED***);
  if (state.deopted) {
    node.body.body.unshift(loop);
  ***REMOVED*** else {
    let target = path.getEarliestCommonAncestorFrom(state.references).getStatementParent();
    target.findParent(path => {
      if (path.isLoop()) {
        target = path;
      ***REMOVED*** else {
        return path.isFunction();
      ***REMOVED***
    ***REMOVED***);
    target.insertBefore(loop);
  ***REMOVED***
  return true;
***REMOVED***
