// ajv/dist/vocabularies/applicator/contains.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const codegen_1 = require("../../compile/codegen");
const util_1 = require("../../compile/util");
const error = {
    message: ({ params: { min, max ***REMOVED*** ***REMOVED***) => max === undefined
        ? (0, codegen_1.str) `must contain at least ${min***REMOVED*** valid item(s)`
        : (0, codegen_1.str) `must contain at least ${min***REMOVED*** and no more than ${max***REMOVED*** valid item(s)`,
    params: ({ params: { min, max ***REMOVED*** ***REMOVED***) => max === undefined ? (0, codegen_1._) `{minContains: ${min***REMOVED******REMOVED***` : (0, codegen_1._) `{minContains: ${min***REMOVED***, maxContains: ${max***REMOVED******REMOVED***`,
***REMOVED***;
const def = {
    keyword: "contains",
    type: "array",
    schemaType: ["object", "boolean"],
    before: "uniqueItems",
    trackErrors: true,
    error,
    code(cxt) {
        const { gen, schema, parentSchema, data, it ***REMOVED*** = cxt;
        let min;
        let max;
        const { minContains, maxContains ***REMOVED*** = parentSchema;
        if (it.opts.next) {
            min = minContains === undefined ? 1 : minContains;
            max = maxContains;
        ***REMOVED***
        else {
            min = 1;
        ***REMOVED***
        const len = gen.const("len", (0, codegen_1._) `${data***REMOVED***.length`);
        cxt.setParams({ min, max ***REMOVED***);
        if (max === undefined && min === 0) {
            (0, util_1.checkStrictMode)(it, `"minContains" == 0 without "maxContains": "contains" keyword ignored`);
            return;
        ***REMOVED***
        if (max !== undefined && min > max) {
            (0, util_1.checkStrictMode)(it, `"minContains" > "maxContains" is always invalid`);
            cxt.fail();
            return;
        ***REMOVED***
        if ((0, util_1.alwaysValidSchema)(it, schema)) {
            let cond = (0, codegen_1._) `${len***REMOVED*** >= ${min***REMOVED***`;
            if (max !== undefined)
                cond = (0, codegen_1._) `${cond***REMOVED*** && ${len***REMOVED*** <= ${max***REMOVED***`;
            cxt.pass(cond);
            return;
        ***REMOVED***
        it.items = true;
        const valid = gen.name("valid");
        if (max === undefined && min === 1) {
            validateItems(valid, () => gen.if(valid, () => gen.break()));
        ***REMOVED***
        else if (min === 0) {
            gen.let(valid, true);
            if (max !== undefined)
                gen.if((0, codegen_1._) `${data***REMOVED***.length > 0`, validateItemsWithCount);
        ***REMOVED***
        else {
            gen.let(valid, false);
            validateItemsWithCount();
        ***REMOVED***
        cxt.result(valid, () => cxt.reset());
        function validateItemsWithCount() {
            const schValid = gen.name("_valid");
            const count = gen.let("count", 0);
            validateItems(schValid, () => gen.if(schValid, () => checkLimits(count)));
        ***REMOVED***
        function validateItems(_valid, block) {
            gen.forRange("i", 0, len, (i) => {
                cxt.subschema({
                    keyword: "contains",
                    dataProp: i,
                    dataPropType: util_1.Type.Num,
                    compositeRule: true,
                ***REMOVED***, _valid);
                block();
            ***REMOVED***);
        ***REMOVED***
        function checkLimits(count) {
            gen.code((0, codegen_1._) `${count***REMOVED***++`);
            if (max === undefined) {
                gen.if((0, codegen_1._) `${count***REMOVED*** >= ${min***REMOVED***`, () => gen.assign(valid, true).break());
            ***REMOVED***
            else {
                gen.if((0, codegen_1._) `${count***REMOVED*** > ${max***REMOVED***`, () => gen.assign(valid, false).break());
                if (min === 1)
                    gen.assign(valid, true);
                else
                    gen.if((0, codegen_1._) `${count***REMOVED*** >= ${min***REMOVED***`, () => gen.assign(valid, true));
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***,
***REMOVED***;
exports.default = def;
