// ajv/dist/vocabularies/applicator/dependencies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.validateSchemaDeps = exports.validatePropertyDeps = exports.error = void 0;
const codegen_1 = require("../../compile/codegen");
const util_1 = require("../../compile/util");
const code_1 = require("../code");
exports.error = {
    message: ({ params: { property, depsCount, deps ***REMOVED*** ***REMOVED***) => {
        const property_ies = depsCount === 1 ? "property" : "properties";
        return (0, codegen_1.str) `must have ${property_ies***REMOVED*** ${deps***REMOVED*** when property ${property***REMOVED*** is present`;
    ***REMOVED***,
    params: ({ params: { property, depsCount, deps, missingProperty ***REMOVED*** ***REMOVED***) => (0, codegen_1._) `{property: ${property***REMOVED***,
    missingProperty: ${missingProperty***REMOVED***,
    depsCount: ${depsCount***REMOVED***,
    deps: ${deps***REMOVED******REMOVED***`, // TODO change to reference
***REMOVED***;
const def = {
    keyword: "dependencies",
    type: "object",
    schemaType: "object",
    error: exports.error,
    code(cxt) {
        const [propDeps, schDeps] = splitDependencies(cxt);
        validatePropertyDeps(cxt, propDeps);
        validateSchemaDeps(cxt, schDeps);
    ***REMOVED***,
***REMOVED***;
function splitDependencies({ schema ***REMOVED***) {
    const propertyDeps = {***REMOVED***;
    const schemaDeps = {***REMOVED***;
    for (const key in schema) {
        if (key === "__proto__")
            continue;
        const deps = Array.isArray(schema[key]) ? propertyDeps : schemaDeps;
        deps[key] = schema[key];
    ***REMOVED***
    return [propertyDeps, schemaDeps];
***REMOVED***
function validatePropertyDeps(cxt, propertyDeps = cxt.schema) {
    const { gen, data, it ***REMOVED*** = cxt;
    if (Object.keys(propertyDeps).length === 0)
        return;
    const missing = gen.let("missing");
    for (const prop in propertyDeps) {
        const deps = propertyDeps[prop];
        if (deps.length === 0)
            continue;
        const hasProperty = (0, code_1.propertyInData)(gen, data, prop, it.opts.ownProperties);
        cxt.setParams({
            property: prop,
            depsCount: deps.length,
            deps: deps.join(", "),
        ***REMOVED***);
        if (it.allErrors) {
            gen.if(hasProperty, () => {
                for (const depProp of deps) {
                    (0, code_1.checkReportMissingProp)(cxt, depProp);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***
        else {
            gen.if((0, codegen_1._) `${hasProperty***REMOVED*** && (${(0, code_1.checkMissingProp)(cxt, deps, missing)***REMOVED***)`);
            (0, code_1.reportMissingProp)(cxt, missing);
            gen.else();
        ***REMOVED***
    ***REMOVED***
***REMOVED***
exports.validatePropertyDeps = validatePropertyDeps;
function validateSchemaDeps(cxt, schemaDeps = cxt.schema) {
    const { gen, data, keyword, it ***REMOVED*** = cxt;
    const valid = gen.name("valid");
    for (const prop in schemaDeps) {
        if ((0, util_1.alwaysValidSchema)(it, schemaDeps[prop]))
            continue;
        gen.if((0, code_1.propertyInData)(gen, data, prop, it.opts.ownProperties), () => {
            const schCxt = cxt.subschema({ keyword, schemaProp: prop ***REMOVED***, valid);
            cxt.mergeValidEvaluated(schCxt, valid);
        ***REMOVED***, () => gen.var(valid, true) // TODO var
        );
        cxt.ok(valid);
    ***REMOVED***
***REMOVED***
exports.validateSchemaDeps = validateSchemaDeps;
exports.default = def;
