// ajv/dist/vocabularies/applicator/oneOf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const codegen_1 = require("../../compile/codegen");
const util_1 = require("../../compile/util");
const error = {
    message: "must match exactly one schema in oneOf",
    params: ({ params ***REMOVED***) => (0, codegen_1._) `{passingSchemas: ${params.passing***REMOVED******REMOVED***`,
***REMOVED***;
const def = {
    keyword: "oneOf",
    schemaType: "array",
    trackErrors: true,
    error,
    code(cxt) {
        const { gen, schema, parentSchema, it ***REMOVED*** = cxt;
        /* istanbul ignore if */
        if (!Array.isArray(schema))
            throw new Error("ajv implementation error");
        if (it.opts.discriminator && parentSchema.discriminator)
            return;
        const schArr = schema;
        const valid = gen.let("valid", false);
        const passing = gen.let("passing", null);
        const schValid = gen.name("_valid");
        cxt.setParams({ passing ***REMOVED***);
        // TODO possibly fail straight away (with warning or exception) if there are two empty always valid schemas
        gen.block(validateOneOf);
        cxt.result(valid, () => cxt.reset(), () => cxt.error(true));
        function validateOneOf() {
            schArr.forEach((sch, i) => {
                let schCxt;
                if ((0, util_1.alwaysValidSchema)(it, sch)) {
                    gen.var(schValid, true);
                ***REMOVED***
                else {
                    schCxt = cxt.subschema({
                        keyword: "oneOf",
                        schemaProp: i,
                        compositeRule: true,
                    ***REMOVED***, schValid);
                ***REMOVED***
                if (i > 0) {
                    gen
                        .if((0, codegen_1._) `${schValid***REMOVED*** && ${valid***REMOVED***`)
                        .assign(valid, false)
                        .assign(passing, (0, codegen_1._) `[${passing***REMOVED***, ${i***REMOVED***]`)
                        .else();
                ***REMOVED***
                gen.if(schValid, () => {
                    gen.assign(valid, true);
                    gen.assign(passing, i);
                    if (schCxt)
                        cxt.mergeEvaluated(schCxt, codegen_1.Name);
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***,
***REMOVED***;
exports.default = def;
