// stylehacks/src/index.js
'use strict';
const browserslist = require('browserslist');
const plugins = require('./plugins');

/** @typedef {{lint?: boolean***REMOVED******REMOVED*** Options */

/**
 * @type {import('postcss').PluginCreator<Options>***REMOVED***
 * @param {Options***REMOVED*** opts
 * @return {import('postcss').Plugin***REMOVED***
 */
function pluginCreator(opts = {***REMOVED***) {
  return {
    postcssPlugin: 'stylehacks',

    OnceExit(css, { result ***REMOVED***) {
      /** @type {typeof result.opts & browserslist.Options***REMOVED*** */
      const resultOpts = result.opts || {***REMOVED***;
      const browsers = browserslist(null, {
        stats: resultOpts.stats,
        path: __dirname,
        env: resultOpts.env,
      ***REMOVED***);

      /** @type {import('./plugin').Plugin[]***REMOVED*** */
      const processors = [];
      for (const Plugin of plugins) {
        const hack = new Plugin(result);
        if (!browsers.some((browser) => hack.targets.has(browser))) {
          processors.push(hack);
        ***REMOVED***
      ***REMOVED***
      css.walk((node) => {
        processors.forEach((proc) => {
          if (!proc.nodeTypes.has(node.type)) {
            return;
          ***REMOVED***

          if (opts.lint) {
            return proc.detectAndWarn(node);
          ***REMOVED***

          return proc.detectAndResolve(node);
        ***REMOVED***);
      ***REMOVED***);
    ***REMOVED***,
  ***REMOVED***;
***REMOVED***

/** @type {(node: import('postcss').Node) => boolean***REMOVED*** */
pluginCreator.detect = (node) => {
  return plugins.some((Plugin) => {
    const hack = new Plugin();

    return hack.any(node);
  ***REMOVED***);
***REMOVED***;

pluginCreator.postcss = true;
module.exports = pluginCreator;
