// domutils/lib/legacy.js
var ElementType = require("domelementtype");
var isTag = exports.isTag = ElementType.isTag;

exports.testElement = function(options, element){
	for(var key in options){
		if(!options.hasOwnProperty(key));
		else if(key === "tag_name"){
			if(!isTag(element) || !options.tag_name(element.name)){
				return false;
			***REMOVED***
		***REMOVED*** else if(key === "tag_type"){
			if(!options.tag_type(element.type)) return false;
		***REMOVED*** else if(key === "tag_contains"){
			if(isTag(element) || !options.tag_contains(element.data)){
				return false;
			***REMOVED***
		***REMOVED*** else if(!element.attribs || !options[key](element.attribs[key])){
			return false;
		***REMOVED***
	***REMOVED***
	return true;
***REMOVED***;

var Checks = {
	tag_name: function(name){
		if(typeof name === "function"){
			return function(elem){ return isTag(elem) && name(elem.name); ***REMOVED***;
		***REMOVED*** else if(name === "*"){
			return isTag;
		***REMOVED*** else {
			return function(elem){ return isTag(elem) && elem.name === name; ***REMOVED***;
		***REMOVED***
	***REMOVED***,
	tag_type: function(type){
		if(typeof type === "function"){
			return function(elem){ return type(elem.type); ***REMOVED***;
		***REMOVED*** else {
			return function(elem){ return elem.type === type; ***REMOVED***;
		***REMOVED***
	***REMOVED***,
	tag_contains: function(data){
		if(typeof data === "function"){
			return function(elem){ return !isTag(elem) && data(elem.data); ***REMOVED***;
		***REMOVED*** else {
			return function(elem){ return !isTag(elem) && elem.data === data; ***REMOVED***;
		***REMOVED***
	***REMOVED***
***REMOVED***;

function getAttribCheck(attrib, value){
	if(typeof value === "function"){
		return function(elem){ return elem.attribs && value(elem.attribs[attrib]); ***REMOVED***;
	***REMOVED*** else {
		return function(elem){ return elem.attribs && elem.attribs[attrib] === value; ***REMOVED***;
	***REMOVED***
***REMOVED***

function combineFuncs(a, b){
	return function(elem){
		return a(elem) || b(elem);
	***REMOVED***;
***REMOVED***

exports.getElements = function(options, element, recurse, limit){
	var funcs = Object.keys(options).map(function(key){
		var value = options[key];
		return key in Checks ? Checks[key](value) : getAttribCheck(key, value);
	***REMOVED***);

	return funcs.length === 0 ? [] : this.filter(
		funcs.reduce(combineFuncs),
		element, recurse, limit
	);
***REMOVED***;

exports.getElementById = function(id, element, recurse){
	if(!Array.isArray(element)) element = [element];
	return this.findOne(getAttribCheck("id", id), element, recurse !== false);
***REMOVED***;

exports.getElementsByTagName = function(name, element, recurse, limit){
	return this.filter(Checks.tag_name(name), element, recurse, limit);
***REMOVED***;

exports.getElementsByTagType = function(type, element, recurse, limit){
	return this.filter(Checks.tag_type(type), element, recurse, limit);
***REMOVED***;
