// webpack-manifest-plugin/dist/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.transformFiles = exports.reduceChunk = exports.reduceAssets = exports.generateManifest = void 0;
const path_1 = require("path");
const generateManifest = (compilation, files, { generate, seed = {***REMOVED*** ***REMOVED***) => {
    let result;
    if (generate) {
        const entrypointsArray = Array.from(compilation.entrypoints.entries());
        const entrypoints = entrypointsArray.reduce((e, [name, entrypoint]) => Object.assign(e, { [name]: entrypoint.getFiles() ***REMOVED***), {***REMOVED***);
        result = generate(seed, files, entrypoints);
    ***REMOVED***
    else {
        result = files.reduce((manifest, file) => Object.assign(manifest, { [file.name]: file.path ***REMOVED***), seed);
    ***REMOVED***
    return result;
***REMOVED***;
exports.generateManifest = generateManifest;
const getFileType = (fileName, { transformExtensions ***REMOVED***) => {
    const replaced = fileName.replace(/\?.*/, '');
    const split = replaced.split('.');
    const extension = split.pop();
    return transformExtensions.test(extension) ? `${split.pop()***REMOVED***.${extension***REMOVED***` : extension;
***REMOVED***;
const reduceAssets = (files, asset, moduleAssets) => {
    let name;
    if (moduleAssets[asset.name]) {
        name = moduleAssets[asset.name];
    ***REMOVED***
    else if (asset.info.sourceFilename) {
        name = path_1.join(path_1.dirname(asset.name), path_1.basename(asset.info.sourceFilename));
    ***REMOVED***
    if (name) {
        return files.concat({
            isAsset: true,
            isChunk: false,
            isInitial: false,
            isModuleAsset: true,
            name,
            path: asset.name
        ***REMOVED***);
    ***REMOVED***
    const isEntryAsset = asset.chunks && asset.chunks.length > 0;
    if (isEntryAsset) {
        return files;
    ***REMOVED***
    return files.concat({
        isAsset: true,
        isChunk: false,
        isInitial: false,
        isModuleAsset: false,
        name: asset.name,
        path: asset.name
    ***REMOVED***);
***REMOVED***;
exports.reduceAssets = reduceAssets;
const reduceChunk = (files, chunk, options, auxiliaryFiles) => {
    Array.from(chunk.auxiliaryFiles || []).forEach((auxiliaryFile) => {
        auxiliaryFiles[auxiliaryFile] = {
            isAsset: true,
            isChunk: false,
            isInitial: false,
            isModuleAsset: true,
            name: path_1.basename(auxiliaryFile),
            path: auxiliaryFile
        ***REMOVED***;
    ***REMOVED***);
    return Array.from(chunk.files).reduce((prev, path) => {
        let name = chunk.name ? chunk.name : null;
        name = name
            ? options.useEntryKeys && !path.endsWith('.map')
                ? name
                : `${name***REMOVED***.${getFileType(path, options)***REMOVED***`
            : path;
        return prev.concat({
            chunk,
            isAsset: false,
            isChunk: true,
            isInitial: chunk.isOnlyInitial(),
            isModuleAsset: false,
            name,
            path
        ***REMOVED***);
    ***REMOVED***, files);
***REMOVED***;
exports.reduceChunk = reduceChunk;
const standardizeFilePaths = (file) => {
    const result = Object.assign({***REMOVED***, file);
    result.name = file.name.replace(/\\/g, '/');
    result.path = file.path.replace(/\\/g, '/');
    return result;
***REMOVED***;
const transformFiles = (files, options) => ['filter', 'map', 'sort']
    .filter((fname) => !!options[fname])
    .reduce((prev, fname) => prev[fname](options[fname]), files)
    .map(standardizeFilePaths);
exports.transformFiles = transformFiles;
