// webpack-manifest-plugin/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod ***REMOVED***;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.WebpackManifestPlugin = exports.getCompilerHooks = void 0;
const path_1 = require("path");
const webpack_1 = __importDefault(require("webpack"));
const NormalModule_1 = __importDefault(require("webpack/lib/NormalModule"));
const hooks_1 = require("./hooks");
Object.defineProperty(exports, "getCompilerHooks", { enumerable: true, get: function () { return hooks_1.getCompilerHooks; ***REMOVED*** ***REMOVED***);
const emitCountMap = new Map();
const defaults = {
    assetHookStage: Infinity,
    basePath: '',
    fileName: 'manifest.json',
    filter: null,
    generate: void 0,
    map: null,
    publicPath: null,
    removeKeyHash: /([a-f0-9]{16,32***REMOVED***\.?)/gi,
    seed: void 0,
    serialize(manifest) {
        return JSON.stringify(manifest, null, 2);
    ***REMOVED***,
    sort: null,
    transformExtensions: /^(gz|map)$/i,
    useEntryKeys: false,
    useLegacyEmit: false,
    writeToFileEmit: false
***REMOVED***;
class WebpackManifestPlugin {
    constructor(opts) {
        this.options = Object.assign({***REMOVED***, defaults, opts);
    ***REMOVED***
    apply(compiler) {
        var _a, _b, _c;
        const moduleAssets = {***REMOVED***;
        const manifestFileName = path_1.resolve(((_a = compiler.options.output) === null || _a === void 0 ? void 0 : _a.path) || './', this.options.fileName);
        const manifestAssetId = path_1.relative(((_b = compiler.options.output) === null || _b === void 0 ? void 0 : _b.path) || './', manifestFileName);
        const beforeRun = hooks_1.beforeRunHook.bind(this, { emitCountMap, manifestFileName ***REMOVED***);
        const emit = hooks_1.emitHook.bind(this, {
            compiler,
            emitCountMap,
            manifestAssetId,
            manifestFileName,
            moduleAssets,
            options: this.options
        ***REMOVED***);
        const normalModuleLoader = hooks_1.normalModuleLoaderHook.bind(this, { moduleAssets ***REMOVED***);
        const hookOptions = {
            name: 'WebpackManifestPlugin',
            stage: this.options.assetHookStage
        ***REMOVED***;
        compiler.hooks.compilation.tap(hookOptions, (compilation) => {
            const hook = !NormalModule_1.default.getCompilationHooks
                ? compilation.hooks.normalModuleLoader
                : NormalModule_1.default.getCompilationHooks(compilation).loader;
            hook.tap(hookOptions, normalModuleLoader);
        ***REMOVED***);
        if (((_c = webpack_1.default.version) === null || _c === void 0 ? void 0 : _c.startsWith('4')) || this.options.useLegacyEmit === true) {
            compiler.hooks.emit.tap(hookOptions, emit);
        ***REMOVED***
        else {
            compiler.hooks.thisCompilation.tap(hookOptions, (compilation) => {
                compilation.hooks.processAssets.tap(hookOptions, () => emit(compilation));
            ***REMOVED***);
        ***REMOVED***
        compiler.hooks.run.tap(hookOptions, beforeRun);
        compiler.hooks.watchRun.tap(hookOptions, beforeRun);
    ***REMOVED***
***REMOVED***
exports.WebpackManifestPlugin = WebpackManifestPlugin;
