// eslint/lib/cli-engine/file-enumerator.js
/**
 * @fileoverview `FileEnumerator` class.
 *
 * `FileEnumerator` class has two responsibilities:
 *
 * 1. Find target files by processing glob patterns.
 * 2. Tie each target file and appropriate configuration.
 *
 * It provides a method:
 *
 * - `iterateFiles(patterns)`
 *     Iterate files which are matched by given patterns together with the
 *     corresponded configuration. This is for `CLIEngine#executeOnFiles()`.
 *     While iterating files, it loads the configuration file of each directory
 *     before iterate files on the directory, so we can use the configuration
 *     files to determine target files.
 *
 * @example
 * const enumerator = new FileEnumerator();
 * const linter = new Linter();
 *
 * for (const { config, filePath ***REMOVED*** of enumerator.iterateFiles(["*.js"])) {
 *     const code = fs.readFileSync(filePath, "utf8");
 *     const messages = linter.verify(code, config, filePath);
 *
 *     console.log(messages);
 * ***REMOVED***
 *
 * @author Toru Nagashima <https://github.com/mysticatea>
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const fs = require("fs");
const path = require("path");
const getGlobParent = require("glob-parent");
const isGlob = require("is-glob");
const escapeRegExp = require("escape-string-regexp");
const { Minimatch ***REMOVED*** = require("minimatch");

const {
    Legacy: {
        IgnorePattern,
        CascadingConfigArrayFactory
    ***REMOVED***
***REMOVED*** = require("@eslint/eslintrc");
const debug = require("debug")("eslint:file-enumerator");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

const minimatchOpts = { dot: true, matchBase: true ***REMOVED***;
const dotfilesPattern = /(?:(?:^\.)|(?:[/\\]\.))[^/\\.].*/u;
const NONE = 0;
const IGNORED_SILENTLY = 1;
const IGNORED = 2;

// For VSCode intellisense
/** @typedef {ReturnType<CascadingConfigArrayFactory["getConfigArrayForFile"]>***REMOVED*** ConfigArray */

/**
 * @typedef {Object***REMOVED*** FileEnumeratorOptions
 * @property {CascadingConfigArrayFactory***REMOVED*** [configArrayFactory] The factory for config arrays.
 * @property {string***REMOVED*** [cwd] The base directory to start lookup.
 * @property {string[]***REMOVED*** [extensions] The extensions to match files for directory patterns.
 * @property {boolean***REMOVED*** [globInputPaths] Set to false to skip glob resolution of input file paths to lint (default: true). If false, each input file paths is assumed to be a non-glob path to an existing file.
 * @property {boolean***REMOVED*** [ignore] The flag to check ignored files.
 * @property {string[]***REMOVED*** [rulePaths] The value of `--rulesdir` option.
 */

/**
 * @typedef {Object***REMOVED*** FileAndConfig
 * @property {string***REMOVED*** filePath The path to a target file.
 * @property {ConfigArray***REMOVED*** config The config entries of that file.
 * @property {boolean***REMOVED*** ignored If `true` then this file should be ignored and warned because it was directly specified.
 */

/**
 * @typedef {Object***REMOVED*** FileEntry
 * @property {string***REMOVED*** filePath The path to a target file.
 * @property {ConfigArray***REMOVED*** config The config entries of that file.
 * @property {NONE|IGNORED_SILENTLY|IGNORED***REMOVED*** flag The flag.
 * - `NONE` means the file is a target file.
 * - `IGNORED_SILENTLY` means the file should be ignored silently.
 * - `IGNORED` means the file should be ignored and warned because it was directly specified.
 */

/**
 * @typedef {Object***REMOVED*** FileEnumeratorInternalSlots
 * @property {CascadingConfigArrayFactory***REMOVED*** configArrayFactory The factory for config arrays.
 * @property {string***REMOVED*** cwd The base directory to start lookup.
 * @property {RegExp|null***REMOVED*** extensionRegExp The RegExp to test if a string ends with specific file extensions.
 * @property {boolean***REMOVED*** globInputPaths Set to false to skip glob resolution of input file paths to lint (default: true). If false, each input file paths is assumed to be a non-glob path to an existing file.
 * @property {boolean***REMOVED*** ignoreFlag The flag to check ignored files.
 * @property {(filePath:string, dot:boolean) => boolean***REMOVED*** defaultIgnores The default predicate function to ignore files.
 */

/** @type {WeakMap<FileEnumerator, FileEnumeratorInternalSlots>***REMOVED*** */
const internalSlotsMap = new WeakMap();

/**
 * Check if a string is a glob pattern or not.
 * @param {string***REMOVED*** pattern A glob pattern.
 * @returns {boolean***REMOVED*** `true` if the string is a glob pattern.
 */
function isGlobPattern(pattern) {
    return isGlob(path.sep === "\\" ? pattern.replace(/\\/gu, "/") : pattern);
***REMOVED***

/**
 * Get stats of a given path.
 * @param {string***REMOVED*** filePath The path to target file.
 * @returns {fs.Stats|null***REMOVED*** The stats.
 * @private
 */
function statSafeSync(filePath) {
    try {
        return fs.statSync(filePath);
    ***REMOVED*** catch (error) {
        /* istanbul ignore next */
        if (error.code !== "ENOENT") {
            throw error;
        ***REMOVED***
        return null;
    ***REMOVED***
***REMOVED***

/**
 * Get filenames in a given path to a directory.
 * @param {string***REMOVED*** directoryPath The path to target directory.
 * @returns {import("fs").Dirent[]***REMOVED*** The filenames.
 * @private
 */
function readdirSafeSync(directoryPath) {
    try {
        return fs.readdirSync(directoryPath, { withFileTypes: true ***REMOVED***);
    ***REMOVED*** catch (error) {
        /* istanbul ignore next */
        if (error.code !== "ENOENT") {
            throw error;
        ***REMOVED***
        return [];
    ***REMOVED***
***REMOVED***

/**
 * Create a `RegExp` object to detect extensions.
 * @param {string[] | null***REMOVED*** extensions The extensions to create.
 * @returns {RegExp | null***REMOVED*** The created `RegExp` object or null.
 */
function createExtensionRegExp(extensions) {
    if (extensions) {
        const normalizedExts = extensions.map(ext => escapeRegExp(
            ext.startsWith(".")
                ? ext.slice(1)
                : ext
        ));

        return new RegExp(
            `.\\.(?:${normalizedExts.join("|")***REMOVED***)$`,
            "u"
        );
    ***REMOVED***
    return null;
***REMOVED***

/**
 * The error type when no files match a glob.
 */
class NoFilesFoundError extends Error {

    // eslint-disable-next-line jsdoc/require-description
    /**
     * @param {string***REMOVED*** pattern The glob pattern which was not found.
     * @param {boolean***REMOVED*** globDisabled If `true` then the pattern was a glob pattern, but glob was disabled.
     */
    constructor(pattern, globDisabled) {
        super(`No files matching '${pattern***REMOVED***' were found${globDisabled ? " (glob was disabled)" : ""***REMOVED***.`);
        this.messageTemplate = "file-not-found";
        this.messageData = { pattern, globDisabled ***REMOVED***;
    ***REMOVED***
***REMOVED***

/**
 * The error type when there are files matched by a glob, but all of them have been ignored.
 */
class AllFilesIgnoredError extends Error {

    // eslint-disable-next-line jsdoc/require-description
    /**
     * @param {string***REMOVED*** pattern The glob pattern which was not found.
     */
    constructor(pattern) {
        super(`All files matched by '${pattern***REMOVED***' are ignored.`);
        this.messageTemplate = "all-files-ignored";
        this.messageData = { pattern ***REMOVED***;
    ***REMOVED***
***REMOVED***

/**
 * This class provides the functionality that enumerates every file which is
 * matched by given glob patterns and that configuration.
 */
class FileEnumerator {

    /**
     * Initialize this enumerator.
     * @param {FileEnumeratorOptions***REMOVED*** options The options.
     */
    constructor({
        cwd = process.cwd(),
        configArrayFactory = new CascadingConfigArrayFactory({
            cwd,
            eslintRecommendedPath: path.resolve(__dirname, "../../conf/eslint-recommended.js"),
            eslintAllPath: path.resolve(__dirname, "../../conf/eslint-all.js")
        ***REMOVED***),
        extensions = null,
        globInputPaths = true,
        errorOnUnmatchedPattern = true,
        ignore = true
    ***REMOVED*** = {***REMOVED***) {
        internalSlotsMap.set(this, {
            configArrayFactory,
            cwd,
            defaultIgnores: IgnorePattern.createDefaultIgnore(cwd),
            extensionRegExp: createExtensionRegExp(extensions),
            globInputPaths,
            errorOnUnmatchedPattern,
            ignoreFlag: ignore
        ***REMOVED***);
    ***REMOVED***

    /**
     * Check if a given file is target or not.
     * @param {string***REMOVED*** filePath The path to a candidate file.
     * @param {ConfigArray***REMOVED*** [providedConfig] Optional. The configuration for the file.
     * @returns {boolean***REMOVED*** `true` if the file is a target.
     */
    isTargetPath(filePath, providedConfig) {
        const {
            configArrayFactory,
            extensionRegExp
        ***REMOVED*** = internalSlotsMap.get(this);

        // If `--ext` option is present, use it.
        if (extensionRegExp) {
            return extensionRegExp.test(filePath);
        ***REMOVED***

        // `.js` file is target by default.
        if (filePath.endsWith(".js")) {
            return true;
        ***REMOVED***

        // use `overrides[].files` to check additional targets.
        const config =
            providedConfig ||
            configArrayFactory.getConfigArrayForFile(
                filePath,
                { ignoreNotFoundError: true ***REMOVED***
            );

        return config.isAdditionalTargetPath(filePath);
    ***REMOVED***

    /**
     * Iterate files which are matched by given glob patterns.
     * @param {string|string[]***REMOVED*** patternOrPatterns The glob patterns to iterate files.
     * @returns {IterableIterator<FileAndConfig>***REMOVED*** The found files.
     */
    *iterateFiles(patternOrPatterns) {
        const { globInputPaths, errorOnUnmatchedPattern ***REMOVED*** = internalSlotsMap.get(this);
        const patterns = Array.isArray(patternOrPatterns)
            ? patternOrPatterns
            : [patternOrPatterns];

        debug("Start to iterate files: %o", patterns);

        // The set of paths to remove duplicate.
        const set = new Set();

        for (const pattern of patterns) {
            let foundRegardlessOfIgnored = false;
            let found = false;

            // Skip empty string.
            if (!pattern) {
                continue;
            ***REMOVED***

            // Iterate files of this pattern.
            for (const { config, filePath, flag ***REMOVED*** of this._iterateFiles(pattern)) {
                foundRegardlessOfIgnored = true;
                if (flag === IGNORED_SILENTLY) {
                    continue;
                ***REMOVED***
                found = true;

                // Remove duplicate paths while yielding paths.
                if (!set.has(filePath)) {
                    set.add(filePath);
                    yield {
                        config,
                        filePath,
                        ignored: flag === IGNORED
                    ***REMOVED***;
                ***REMOVED***
            ***REMOVED***

            // Raise an error if any files were not found.
            if (errorOnUnmatchedPattern) {
                if (!foundRegardlessOfIgnored) {
                    throw new NoFilesFoundError(
                        pattern,
                        !globInputPaths && isGlob(pattern)
                    );
                ***REMOVED***
                if (!found) {
                    throw new AllFilesIgnoredError(pattern);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        debug(`Complete iterating files: ${JSON.stringify(patterns)***REMOVED***`);
    ***REMOVED***

    /**
     * Iterate files which are matched by a given glob pattern.
     * @param {string***REMOVED*** pattern The glob pattern to iterate files.
     * @returns {IterableIterator<FileEntry>***REMOVED*** The found files.
     */
    _iterateFiles(pattern) {
        const { cwd, globInputPaths ***REMOVED*** = internalSlotsMap.get(this);
        const absolutePath = path.resolve(cwd, pattern);
        const isDot = dotfilesPattern.test(pattern);
        const stat = statSafeSync(absolutePath);

        if (stat && stat.isDirectory()) {
            return this._iterateFilesWithDirectory(absolutePath, isDot);
        ***REMOVED***
        if (stat && stat.isFile()) {
            return this._iterateFilesWithFile(absolutePath);
        ***REMOVED***
        if (globInputPaths && isGlobPattern(pattern)) {
            return this._iterateFilesWithGlob(absolutePath, isDot);
        ***REMOVED***

        return [];
    ***REMOVED***

    /**
     * Iterate a file which is matched by a given path.
     * @param {string***REMOVED*** filePath The path to the target file.
     * @returns {IterableIterator<FileEntry>***REMOVED*** The found files.
     * @private
     */
    _iterateFilesWithFile(filePath) {
        debug(`File: ${filePath***REMOVED***`);

        const { configArrayFactory ***REMOVED*** = internalSlotsMap.get(this);
        const config = configArrayFactory.getConfigArrayForFile(filePath);
        const ignored = this._isIgnoredFile(filePath, { config, direct: true ***REMOVED***);
        const flag = ignored ? IGNORED : NONE;

        return [{ config, filePath, flag ***REMOVED***];
    ***REMOVED***

    /**
     * Iterate files in a given path.
     * @param {string***REMOVED*** directoryPath The path to the target directory.
     * @param {boolean***REMOVED*** dotfiles If `true` then it doesn't skip dot files by default.
     * @returns {IterableIterator<FileEntry>***REMOVED*** The found files.
     * @private
     */
    _iterateFilesWithDirectory(directoryPath, dotfiles) {
        debug(`Directory: ${directoryPath***REMOVED***`);

        return this._iterateFilesRecursive(
            directoryPath,
            { dotfiles, recursive: true, selector: null ***REMOVED***
        );
    ***REMOVED***

    /**
     * Iterate files which are matched by a given glob pattern.
     * @param {string***REMOVED*** pattern The glob pattern to iterate files.
     * @param {boolean***REMOVED*** dotfiles If `true` then it doesn't skip dot files by default.
     * @returns {IterableIterator<FileEntry>***REMOVED*** The found files.
     * @private
     */
    _iterateFilesWithGlob(pattern, dotfiles) {
        debug(`Glob: ${pattern***REMOVED***`);

        const directoryPath = path.resolve(getGlobParent(pattern));
        const globPart = pattern.slice(directoryPath.length + 1);

        /*
         * recursive if there are `**` or path separators in the glob part.
         * Otherwise, patterns such as `src/*.js`, it doesn't need recursive.
         */
        const recursive = /\*\*|\/|\\/u.test(globPart);
        const selector = new Minimatch(pattern, minimatchOpts);

        debug(`recursive? ${recursive***REMOVED***`);

        return this._iterateFilesRecursive(
            directoryPath,
            { dotfiles, recursive, selector ***REMOVED***
        );
    ***REMOVED***

    /**
     * Iterate files in a given path.
     * @param {string***REMOVED*** directoryPath The path to the target directory.
     * @param {Object***REMOVED*** options The options to iterate files.
     * @param {boolean***REMOVED*** [options.dotfiles] If `true` then it doesn't skip dot files by default.
     * @param {boolean***REMOVED*** [options.recursive] If `true` then it dives into sub directories.
     * @param {InstanceType<Minimatch>***REMOVED*** [options.selector] The matcher to choose files.
     * @returns {IterableIterator<FileEntry>***REMOVED*** The found files.
     * @private
     */
    *_iterateFilesRecursive(directoryPath, options) {
        debug(`Enter the directory: ${directoryPath***REMOVED***`);
        const { configArrayFactory ***REMOVED*** = internalSlotsMap.get(this);

        /** @type {ConfigArray|null***REMOVED*** */
        let config = null;

        // Enumerate the files of this directory.
        for (const entry of readdirSafeSync(directoryPath)) {
            const filePath = path.join(directoryPath, entry.name);
            const fileInfo = entry.isSymbolicLink() ? statSafeSync(filePath) : entry;

            if (!fileInfo) {
                continue;
            ***REMOVED***

            // Check if the file is matched.
            if (fileInfo.isFile()) {
                if (!config) {
                    config = configArrayFactory.getConfigArrayForFile(
                        filePath,

                        /*
                         * We must ignore `ConfigurationNotFoundError` at this
                         * point because we don't know if target files exist in
                         * this directory.
                         */
                        { ignoreNotFoundError: true ***REMOVED***
                    );
                ***REMOVED***
                const matched = options.selector

                    // Started with a glob pattern; choose by the pattern.
                    ? options.selector.match(filePath)

                    // Started with a directory path; choose by file extensions.
                    : this.isTargetPath(filePath, config);

                if (matched) {
                    const ignored = this._isIgnoredFile(filePath, { ...options, config ***REMOVED***);
                    const flag = ignored ? IGNORED_SILENTLY : NONE;

                    debug(`Yield: ${entry.name***REMOVED***${ignored ? " but ignored" : ""***REMOVED***`);
                    yield {
                        config: configArrayFactory.getConfigArrayForFile(filePath),
                        filePath,
                        flag
                    ***REMOVED***;
                ***REMOVED*** else {
                    debug(`Didn't match: ${entry.name***REMOVED***`);
                ***REMOVED***

            // Dive into the sub directory.
            ***REMOVED*** else if (options.recursive && fileInfo.isDirectory()) {
                if (!config) {
                    config = configArrayFactory.getConfigArrayForFile(
                        filePath,
                        { ignoreNotFoundError: true ***REMOVED***
                    );
                ***REMOVED***
                const ignored = this._isIgnoredFile(
                    filePath + path.sep,
                    { ...options, config ***REMOVED***
                );

                if (!ignored) {
                    yield* this._iterateFilesRecursive(filePath, options);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        debug(`Leave the directory: ${directoryPath***REMOVED***`);
    ***REMOVED***

    /**
     * Check if a given file should be ignored.
     * @param {string***REMOVED*** filePath The path to a file to check.
     * @param {Object***REMOVED*** options Options
     * @param {ConfigArray***REMOVED*** [options.config] The config for this file.
     * @param {boolean***REMOVED*** [options.dotfiles] If `true` then this is not ignore dot files by default.
     * @param {boolean***REMOVED*** [options.direct] If `true` then this is a direct specified file.
     * @returns {boolean***REMOVED*** `true` if the file should be ignored.
     * @private
     */
    _isIgnoredFile(filePath, {
        config: providedConfig,
        dotfiles = false,
        direct = false
    ***REMOVED***) {
        const {
            configArrayFactory,
            defaultIgnores,
            ignoreFlag
        ***REMOVED*** = internalSlotsMap.get(this);

        if (ignoreFlag) {
            const config =
                providedConfig ||
                configArrayFactory.getConfigArrayForFile(
                    filePath,
                    { ignoreNotFoundError: true ***REMOVED***
                );
            const ignores =
                config.extractConfig(filePath).ignores || defaultIgnores;

            return ignores(filePath, dotfiles);
        ***REMOVED***

        return !direct && defaultIgnores(filePath, dotfiles);
    ***REMOVED***
***REMOVED***

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

module.exports = { FileEnumerator ***REMOVED***;
