// ajv-formats/src/index.ts
import {
  DefinedFormats,
  FormatMode,
  FormatName,
  formatNames,
  fastFormats,
  fullFormats,
***REMOVED*** from "./formats"
import formatLimit from "./limit"
import type Ajv from "ajv"
import type {Plugin, Format***REMOVED*** from "ajv"
import {_, Name***REMOVED*** from "ajv/dist/compile/codegen"

export {FormatMode, FormatName***REMOVED*** from "./formats"
export {LimitFormatError***REMOVED*** from "./limit"
export interface FormatOptions {
  mode?: FormatMode
  formats?: FormatName[]
  keywords?: boolean
***REMOVED***

export type FormatsPluginOptions = FormatName[] | FormatOptions

export interface FormatsPlugin extends Plugin<FormatsPluginOptions> {
  get: (format: FormatName, mode?: FormatMode) => Format
***REMOVED***

const fullName = new Name("fullFormats")
const fastName = new Name("fastFormats")

const formatsPlugin: FormatsPlugin = (
  ajv: Ajv,
  opts: FormatsPluginOptions = {keywords: true***REMOVED***
): Ajv => {
  if (Array.isArray(opts)) {
    addFormats(ajv, opts, fullFormats, fullName)
    return ajv
  ***REMOVED***
  const [formats, exportName] =
    opts.mode === "fast" ? [fastFormats, fastName] : [fullFormats, fullName]
  const list = opts.formats || formatNames
  addFormats(ajv, list, formats, exportName)
  if (opts.keywords) formatLimit(ajv)
  return ajv
***REMOVED***

formatsPlugin.get = (name: FormatName, mode: FormatMode = "full"): Format => {
  const formats = mode === "fast" ? fastFormats : fullFormats
  const f = formats[name]
  if (!f) throw new Error(`Unknown format "${name***REMOVED***"`)
  return f
***REMOVED***

function addFormats(ajv: Ajv, list: FormatName[], fs: DefinedFormats, exportName: Name): void {
  ajv.opts.code.formats ??= _`require("ajv-formats/dist/formats").${exportName***REMOVED***`
  for (const f of list) ajv.addFormat(f, fs[f])
***REMOVED***

module.exports = exports = formatsPlugin
Object.defineProperty(exports, "__esModule", {value: true***REMOVED***)

export default formatsPlugin
