// ajv-formats/src/limit.ts
import type Ajv from "ajv"
import type {
  Plugin,
  CodeKeywordDefinition,
  KeywordErrorDefinition,
  Code,
  Name,
  ErrorObject,
***REMOVED*** from "ajv"
import type {AddedFormat***REMOVED*** from "ajv/dist/types"
import type {Rule***REMOVED*** from "ajv/dist/compile/rules"
import {KeywordCxt***REMOVED*** from "ajv"
import {_, str, or, getProperty, operators***REMOVED*** from "ajv/dist/compile/codegen"

type Kwd = "formatMaximum" | "formatMinimum" | "formatExclusiveMaximum" | "formatExclusiveMinimum"

type Comparison = "<=" | ">=" | "<" | ">"

const ops = operators

const KWDs: {[K in Kwd]: {okStr: Comparison; ok: Code; fail: Code***REMOVED******REMOVED*** = {
  formatMaximum: {okStr: "<=", ok: ops.LTE, fail: ops.GT***REMOVED***,
  formatMinimum: {okStr: ">=", ok: ops.GTE, fail: ops.LT***REMOVED***,
  formatExclusiveMaximum: {okStr: "<", ok: ops.LT, fail: ops.GTE***REMOVED***,
  formatExclusiveMinimum: {okStr: ">", ok: ops.GT, fail: ops.LTE***REMOVED***,
***REMOVED***

export type LimitFormatError = ErrorObject<Kwd, {limit: string; comparison: Comparison***REMOVED***>

const error: KeywordErrorDefinition = {
  message: ({keyword, schemaCode***REMOVED***) => str`should be ${KWDs[keyword as Kwd].okStr***REMOVED*** ${schemaCode***REMOVED***`,
  params: ({keyword, schemaCode***REMOVED***) =>
    _`{comparison: ${KWDs[keyword as Kwd].okStr***REMOVED***, limit: ${schemaCode***REMOVED******REMOVED***`,
***REMOVED***

export const formatLimitDefinition: CodeKeywordDefinition = {
  keyword: Object.keys(KWDs),
  type: "string",
  schemaType: "string",
  $data: true,
  error,
  code(cxt) {
    const {gen, data, schemaCode, keyword, it***REMOVED*** = cxt
    const {opts, self***REMOVED*** = it
    if (!opts.validateFormats) return

    const fCxt = new KeywordCxt(it, (self.RULES.all.format as Rule).definition, "format")
    if (fCxt.$data) validate$DataFormat()
    else validateFormat()

    function validate$DataFormat(): void {
      const fmts = gen.scopeValue("formats", {
        ref: self.formats,
        code: opts.code.formats,
      ***REMOVED***)
      const fmt = gen.const("fmt", _`${fmts***REMOVED***[${fCxt.schemaCode***REMOVED***]`)
      cxt.fail$data(
        or(
          _`typeof ${fmt***REMOVED*** != "object"`,
          _`${fmt***REMOVED*** instanceof RegExp`,
          _`typeof ${fmt***REMOVED***.compare != "function"`,
          compareCode(fmt)
        )
      )
    ***REMOVED***

    function validateFormat(): void {
      const format = fCxt.schema as string
      const fmtDef: AddedFormat | undefined = self.formats[format]
      if (!fmtDef || fmtDef === true) return
      if (
        typeof fmtDef != "object" ||
        fmtDef instanceof RegExp ||
        typeof fmtDef.compare != "function"
      ) {
        throw new Error(`"${keyword***REMOVED***": format "${format***REMOVED***" does not define "compare" function`)
      ***REMOVED***
      const fmt = gen.scopeValue("formats", {
        key: format,
        ref: fmtDef,
        code: opts.code.formats ? _`${opts.code.formats***REMOVED***${getProperty(format)***REMOVED***` : undefined,
      ***REMOVED***)

      cxt.fail$data(compareCode(fmt))
    ***REMOVED***

    function compareCode(fmt: Name): Code {
      return _`${fmt***REMOVED***.compare(${data***REMOVED***, ${schemaCode***REMOVED***) ${KWDs[keyword as Kwd].fail***REMOVED*** 0`
    ***REMOVED***
  ***REMOVED***,
  dependencies: ["format"],
***REMOVED***

const formatLimitPlugin: Plugin<undefined> = (ajv: Ajv): Ajv => {
  ajv.addKeyword(formatLimitDefinition)
  return ajv
***REMOVED***

export default formatLimitPlugin
