// parse5/lib/tree-adapters/default.js
'use strict';

const { DOCUMENT_MODE ***REMOVED*** = require('../common/html');

//Node construction
exports.createDocument = function() {
    return {
        nodeName: '#document',
        mode: DOCUMENT_MODE.NO_QUIRKS,
        childNodes: []
    ***REMOVED***;
***REMOVED***;

exports.createDocumentFragment = function() {
    return {
        nodeName: '#document-fragment',
        childNodes: []
    ***REMOVED***;
***REMOVED***;

exports.createElement = function(tagName, namespaceURI, attrs) {
    return {
        nodeName: tagName,
        tagName: tagName,
        attrs: attrs,
        namespaceURI: namespaceURI,
        childNodes: [],
        parentNode: null
    ***REMOVED***;
***REMOVED***;

exports.createCommentNode = function(data) {
    return {
        nodeName: '#comment',
        data: data,
        parentNode: null
    ***REMOVED***;
***REMOVED***;

const createTextNode = function(value) {
    return {
        nodeName: '#text',
        value: value,
        parentNode: null
    ***REMOVED***;
***REMOVED***;

//Tree mutation
const appendChild = (exports.appendChild = function(parentNode, newNode) {
    parentNode.childNodes.push(newNode);
    newNode.parentNode = parentNode;
***REMOVED***);

const insertBefore = (exports.insertBefore = function(parentNode, newNode, referenceNode) {
    const insertionIdx = parentNode.childNodes.indexOf(referenceNode);

    parentNode.childNodes.splice(insertionIdx, 0, newNode);
    newNode.parentNode = parentNode;
***REMOVED***);

exports.setTemplateContent = function(templateElement, contentElement) {
    templateElement.content = contentElement;
***REMOVED***;

exports.getTemplateContent = function(templateElement) {
    return templateElement.content;
***REMOVED***;

exports.setDocumentType = function(document, name, publicId, systemId) {
    let doctypeNode = null;

    for (let i = 0; i < document.childNodes.length; i++) {
        if (document.childNodes[i].nodeName === '#documentType') {
            doctypeNode = document.childNodes[i];
            break;
        ***REMOVED***
    ***REMOVED***

    if (doctypeNode) {
        doctypeNode.name = name;
        doctypeNode.publicId = publicId;
        doctypeNode.systemId = systemId;
    ***REMOVED*** else {
        appendChild(document, {
            nodeName: '#documentType',
            name: name,
            publicId: publicId,
            systemId: systemId
        ***REMOVED***);
    ***REMOVED***
***REMOVED***;

exports.setDocumentMode = function(document, mode) {
    document.mode = mode;
***REMOVED***;

exports.getDocumentMode = function(document) {
    return document.mode;
***REMOVED***;

exports.detachNode = function(node) {
    if (node.parentNode) {
        const idx = node.parentNode.childNodes.indexOf(node);

        node.parentNode.childNodes.splice(idx, 1);
        node.parentNode = null;
    ***REMOVED***
***REMOVED***;

exports.insertText = function(parentNode, text) {
    if (parentNode.childNodes.length) {
        const prevNode = parentNode.childNodes[parentNode.childNodes.length - 1];

        if (prevNode.nodeName === '#text') {
            prevNode.value += text;
            return;
        ***REMOVED***
    ***REMOVED***

    appendChild(parentNode, createTextNode(text));
***REMOVED***;

exports.insertTextBefore = function(parentNode, text, referenceNode) {
    const prevNode = parentNode.childNodes[parentNode.childNodes.indexOf(referenceNode) - 1];

    if (prevNode && prevNode.nodeName === '#text') {
        prevNode.value += text;
    ***REMOVED*** else {
        insertBefore(parentNode, createTextNode(text), referenceNode);
    ***REMOVED***
***REMOVED***;

exports.adoptAttributes = function(recipient, attrs) {
    const recipientAttrsMap = [];

    for (let i = 0; i < recipient.attrs.length; i++) {
        recipientAttrsMap.push(recipient.attrs[i].name);
    ***REMOVED***

    for (let j = 0; j < attrs.length; j++) {
        if (recipientAttrsMap.indexOf(attrs[j].name) === -1) {
            recipient.attrs.push(attrs[j]);
        ***REMOVED***
    ***REMOVED***
***REMOVED***;

//Tree traversing
exports.getFirstChild = function(node) {
    return node.childNodes[0];
***REMOVED***;

exports.getChildNodes = function(node) {
    return node.childNodes;
***REMOVED***;

exports.getParentNode = function(node) {
    return node.parentNode;
***REMOVED***;

exports.getAttrList = function(element) {
    return element.attrs;
***REMOVED***;

//Node data
exports.getTagName = function(element) {
    return element.tagName;
***REMOVED***;

exports.getNamespaceURI = function(element) {
    return element.namespaceURI;
***REMOVED***;

exports.getTextNodeContent = function(textNode) {
    return textNode.value;
***REMOVED***;

exports.getCommentNodeContent = function(commentNode) {
    return commentNode.data;
***REMOVED***;

exports.getDocumentTypeNodeName = function(doctypeNode) {
    return doctypeNode.name;
***REMOVED***;

exports.getDocumentTypeNodePublicId = function(doctypeNode) {
    return doctypeNode.publicId;
***REMOVED***;

exports.getDocumentTypeNodeSystemId = function(doctypeNode) {
    return doctypeNode.systemId;
***REMOVED***;

//Node types
exports.isTextNode = function(node) {
    return node.nodeName === '#text';
***REMOVED***;

exports.isCommentNode = function(node) {
    return node.nodeName === '#comment';
***REMOVED***;

exports.isDocumentTypeNode = function(node) {
    return node.nodeName === '#documentType';
***REMOVED***;

exports.isElementNode = function(node) {
    return !!node.tagName;
***REMOVED***;

// Source code location
exports.setNodeSourceCodeLocation = function(node, location) {
    node.sourceCodeLocation = location;
***REMOVED***;

exports.getNodeSourceCodeLocation = function(node) {
    return node.sourceCodeLocation;
***REMOVED***;

exports.updateNodeSourceCodeLocation = function(node, endLocation) {
    node.sourceCodeLocation = Object.assign(node.sourceCodeLocation, endLocation);
***REMOVED***;
