// @babel/core/lib/config/helpers/config-api.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.makeConfigAPI = makeConfigAPI;
exports.makePluginAPI = makePluginAPI;
exports.makePresetAPI = makePresetAPI;
function _semver() {
  const data = require("semver");
  _semver = function () {
    return data;
  ***REMOVED***;
  return data;
***REMOVED***
var _index = require("../../index.js");
var _caching = require("../caching.js");
function makeConfigAPI(cache) {
  const env = value => cache.using(data => {
    if (value === undefined) return data.envName;
    if (typeof value === "function") {
      return (0, _caching.assertSimpleType)(value(data.envName));
    ***REMOVED***
    return (Array.isArray(value) ? value : [value]).some(entry => {
      if (typeof entry !== "string") {
        throw new Error("Unexpected non-string value");
      ***REMOVED***
      return entry === data.envName;
    ***REMOVED***);
  ***REMOVED***);
  const caller = cb => cache.using(data => (0, _caching.assertSimpleType)(cb(data.caller)));
  return {
    version: _index.version,
    cache: cache.simple(),
    env,
    async: () => false,
    caller,
    assertVersion
  ***REMOVED***;
***REMOVED***
function makePresetAPI(cache, externalDependencies) {
  const targets = () => JSON.parse(cache.using(data => JSON.stringify(data.targets)));
  const addExternalDependency = ref => {
    externalDependencies.push(ref);
  ***REMOVED***;
  return Object.assign({***REMOVED***, makeConfigAPI(cache), {
    targets,
    addExternalDependency
  ***REMOVED***);
***REMOVED***
function makePluginAPI(cache, externalDependencies) {
  const assumption = name => cache.using(data => data.assumptions[name]);
  return Object.assign({***REMOVED***, makePresetAPI(cache, externalDependencies), {
    assumption
  ***REMOVED***);
***REMOVED***
function assertVersion(range) {
  if (typeof range === "number") {
    if (!Number.isInteger(range)) {
      throw new Error("Expected string or integer value.");
    ***REMOVED***
    range = `^${range***REMOVED***.0.0-0`;
  ***REMOVED***
  if (typeof range !== "string") {
    throw new Error("Expected string or integer value.");
  ***REMOVED***
  if (range === "*" || _semver().satisfies(_index.version, range)) return;
  const limit = Error.stackTraceLimit;
  if (typeof limit === "number" && limit < 25) {
    Error.stackTraceLimit = 25;
  ***REMOVED***
  const err = new Error(`Requires Babel "${range***REMOVED***", but was loaded with "${_index.version***REMOVED***". ` + `If you are sure you have a compatible version of @babel/core, ` + `it is likely that something in your build process is loading the ` + `wrong version. Inspect the stack trace of this error to look for ` + `the first entry that doesn't mention "@babel/core" or "babel-core" ` + `to see what is calling Babel.`);
  if (typeof limit === "number") {
    Error.stackTraceLimit = limit;
  ***REMOVED***
  throw Object.assign(err, {
    code: "BABEL_VERSION_UNSUPPORTED",
    version: _index.version,
    range
  ***REMOVED***);
***REMOVED***
0 && 0;
