// jsdom/lib/jsdom/living/nodes/ChildNode-impl.js
"use strict";

const { convertNodesIntoNode ***REMOVED*** = require("../node");

class ChildNodeImpl {
  remove() {
    if (!this.parentNode) {
      return;
    ***REMOVED***

    this.parentNode._remove(this);
  ***REMOVED***

  after(...nodes) {
    const parent = this.parentNode;
    if (parent) {
      let viableNextSibling = this.nextSibling;
      let idx = viableNextSibling ? nodes.indexOf(viableNextSibling) : -1;

      while (idx !== -1) {
        viableNextSibling = viableNextSibling.nextSibling;
        if (!viableNextSibling) {
          break;
        ***REMOVED***
        idx = nodes.indexOf(viableNextSibling);
      ***REMOVED***

      parent._preInsert(convertNodesIntoNode(this._ownerDocument, nodes), viableNextSibling);
    ***REMOVED***
  ***REMOVED***

  before(...nodes) {
    const parent = this.parentNode;
    if (parent) {
      let viablePreviousSibling = this.previousSibling;
      let idx = viablePreviousSibling ? nodes.indexOf(viablePreviousSibling) : -1;

      while (idx !== -1) {
        viablePreviousSibling = viablePreviousSibling.previousSibling;
        if (!viablePreviousSibling) {
          break;
        ***REMOVED***
        idx = nodes.indexOf(viablePreviousSibling);
      ***REMOVED***

      parent._preInsert(
        convertNodesIntoNode(this._ownerDocument, nodes),
        viablePreviousSibling ? viablePreviousSibling.nextSibling : parent.firstChild
      );
    ***REMOVED***
  ***REMOVED***

  replaceWith(...nodes) {
    const parent = this.parentNode;
    if (parent) {
      let viableNextSibling = this.nextSibling;
      let idx = viableNextSibling ? nodes.indexOf(viableNextSibling) : -1;

      while (idx !== -1) {
        viableNextSibling = viableNextSibling.nextSibling;
        if (!viableNextSibling) {
          break;
        ***REMOVED***
        idx = nodes.indexOf(viableNextSibling);
      ***REMOVED***

      const node = convertNodesIntoNode(this._ownerDocument, nodes);

      if (this.parentNode === parent) {
        parent._replace(node, this);
      ***REMOVED*** else {
        parent._preInsert(node, viableNextSibling);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***

module.exports = {
  implementation: ChildNodeImpl
***REMOVED***;
