// jsdom/lib/jsdom/living/nodes/Document-impl.js
"use strict";

const { CookieJar ***REMOVED*** = require("tough-cookie");

const NodeImpl = require("./Node-impl").implementation;
const idlUtils = require("../generated/utils");
const NODE_TYPE = require("../node-type");
const { hasWeakRefs, mixin, memoizeQuery ***REMOVED*** = require("../../utils");
const { firstChildWithLocalName, firstChildWithLocalNames, firstDescendantWithLocalName ***REMOVED*** =
  require("../helpers/traversal");
const whatwgURL = require("whatwg-url");
const StyleSheetList = require("../generated/StyleSheetList.js");
const { domSymbolTree ***REMOVED*** = require("../helpers/internal-constants");
const eventAccessors = require("../helpers/create-event-accessor");
const { asciiLowercase, stripAndCollapseASCIIWhitespace ***REMOVED*** = require("../helpers/strings");
const { childTextContent ***REMOVED*** = require("../helpers/text");
const { HTML_NS, SVG_NS ***REMOVED*** = require("../helpers/namespaces");
const DOMException = require("domexception/webidl2js-wrapper");
const { parseIntoDocument ***REMOVED*** = require("../../browser/parser");
const History = require("../generated/History");
const Location = require("../generated/Location");
const HTMLCollection = require("../generated/HTMLCollection");
const NodeList = require("../generated/NodeList");
const validateName = require("../helpers/validate-names").name;
const { validateAndExtract ***REMOVED*** = require("../helpers/validate-names");
const { fireAnEvent ***REMOVED*** = require("../helpers/events");
const { shadowIncludingInclusiveDescendantsIterator ***REMOVED*** = require("../helpers/shadow-dom");
const { enqueueCECallbackReaction ***REMOVED*** = require("../helpers/custom-elements");
const { createElement, internalCreateElementNSSteps ***REMOVED*** = require("../helpers/create-element");
const IterableWeakSet = require("../helpers/iterable-weak-set");

const DocumentOrShadowRootImpl = require("./DocumentOrShadowRoot-impl").implementation;
const GlobalEventHandlersImpl = require("./GlobalEventHandlers-impl").implementation;
const NonElementParentNodeImpl = require("./NonElementParentNode-impl").implementation;
const ParentNodeImpl = require("./ParentNode-impl").implementation;

const { clone, listOfElementsWithQualifiedName, listOfElementsWithNamespaceAndLocalName,
  listOfElementsWithClassNames ***REMOVED*** = require("../node");
const generatedAttr = require("../generated/Attr");
const Comment = require("../generated/Comment");
const ProcessingInstruction = require("../generated/ProcessingInstruction");
const CDATASection = require("../generated/CDATASection");
const Text = require("../generated/Text");
const DocumentFragment = require("../generated/DocumentFragment");
const DOMImplementation = require("../generated/DOMImplementation");
const TreeWalker = require("../generated/TreeWalker");
const NodeIterator = require("../generated/NodeIterator");
const ShadowRoot = require("../generated/ShadowRoot");
const Range = require("../generated/Range");
const documents = require("../documents.js");

const CustomEvent = require("../generated/CustomEvent");
const ErrorEvent = require("../generated/ErrorEvent");
const Event = require("../generated/Event");
const FocusEvent = require("../generated/FocusEvent");
const HashChangeEvent = require("../generated/HashChangeEvent");
const KeyboardEvent = require("../generated/KeyboardEvent");
const MessageEvent = require("../generated/MessageEvent");
const MouseEvent = require("../generated/MouseEvent");
const PopStateEvent = require("../generated/PopStateEvent");
const ProgressEvent = require("../generated/ProgressEvent");
const TouchEvent = require("../generated/TouchEvent");
const UIEvent = require("../generated/UIEvent");

const RequestManager = require("../../browser/resources/request-manager");
const AsyncResourceQueue = require("../../browser/resources/async-resource-queue");
const ResourceQueue = require("../../browser/resources/resource-queue");
const PerDocumentResourceLoader = require("../../browser/resources/per-document-resource-loader");

function clearChildNodes(node) {
  for (let child = domSymbolTree.firstChild(node); child; child = domSymbolTree.firstChild(node)) {
    node.removeChild(child);
  ***REMOVED***
***REMOVED***

function pad(number) {
  if (number < 10) {
    return "0" + number;
  ***REMOVED***
  return number;
***REMOVED***

function toLastModifiedString(date) {
  return pad(date.getMonth() + 1) +
    "/" + pad(date.getDate()) +
    "/" + date.getFullYear() +
    " " + pad(date.getHours()) +
    ":" + pad(date.getMinutes()) +
    ":" + pad(date.getSeconds());
***REMOVED***

const eventInterfaceTable = {
  customevent: CustomEvent,
  errorevent: ErrorEvent,
  event: Event,
  events: Event,
  focusevent: FocusEvent,
  hashchangeevent: HashChangeEvent,
  htmlevents: Event,
  keyboardevent: KeyboardEvent,
  messageevent: MessageEvent,
  mouseevent: MouseEvent,
  mouseevents: MouseEvent,
  popstateevent: PopStateEvent,
  progressevent: ProgressEvent,
  svgevents: Event,
  touchevent: TouchEvent,
  uievent: UIEvent,
  uievents: UIEvent
***REMOVED***;

class DocumentImpl extends NodeImpl {
  constructor(globalObject, args, privateData) {
    super(globalObject, args, privateData);

    this._initGlobalEvents();

    this._ownerDocument = this;
    this.nodeType = NODE_TYPE.DOCUMENT_NODE;
    if (!privateData.options) {
      privateData.options = {***REMOVED***;
    ***REMOVED***
    if (!privateData.options.parsingMode) {
      privateData.options.parsingMode = "xml";
    ***REMOVED***
    if (!privateData.options.encoding) {
      privateData.options.encoding = "UTF-8";
    ***REMOVED***
    if (!privateData.options.contentType) {
      privateData.options.contentType = privateData.options.parsingMode === "xml" ? "application/xml" : "text/html";
    ***REMOVED***

    this._parsingMode = privateData.options.parsingMode;

    this._implementation = DOMImplementation.createImpl(this._globalObject, [], {
      ownerDocument: this
    ***REMOVED***);

    this._defaultView = privateData.options.defaultView || null;
    this._global = privateData.options.global;
    this._ids = Object.create(null);
    this._attached = true;
    this._currentScript = null;
    this._pageShowingFlag = false;
    this._cookieJar = privateData.options.cookieJar;
    this._parseOptions = privateData.options.parseOptions || {***REMOVED***;
    this._scriptingDisabled = privateData.options.scriptingDisabled;
    if (this._cookieJar === undefined) {
      this._cookieJar = new CookieJar(null, { looseMode: true ***REMOVED***);
    ***REMOVED***

    if (this._scriptingDisabled) {
      this._parseOptions.scriptingEnabled = false;
    ***REMOVED***

    this.contentType = privateData.options.contentType;
    this._encoding = privateData.options.encoding;

    const urlOption = privateData.options.url === undefined ? "about:blank" : privateData.options.url;
    const parsed = whatwgURL.parseURL(urlOption);
    if (parsed === null) {
      throw new TypeError(`Could not parse "${urlOption***REMOVED***" as a URL`);
    ***REMOVED***

    this._URL = parsed;
    this._origin = urlOption === "about:blank" && privateData.options.parentOrigin ?
      privateData.options.parentOrigin :
      whatwgURL.serializeURLOrigin(this._URL);

    this._location = Location.createImpl(this._globalObject, [], { relevantDocument: this ***REMOVED***);
    this._history = History.createImpl(this._globalObject, [], {
      window: this._defaultView,
      document: this,
      actAsIfLocationReloadCalled: () => this._location.reload()
    ***REMOVED***);

    if (hasWeakRefs) {
      this._workingNodeIterators = new IterableWeakSet();
    ***REMOVED*** else {
      this._workingNodeIterators = [];
    ***REMOVED***

    this._referrer = privateData.options.referrer || "";
    this._lastModified = toLastModifiedString(privateData.options.lastModified || new Date());
    this._asyncQueue = new AsyncResourceQueue();
    this._queue = new ResourceQueue({ asyncQueue: this._asyncQueue, paused: false ***REMOVED***);
    this._deferQueue = new ResourceQueue({ paused: true ***REMOVED***);
    this._requestManager = new RequestManager();
    this._currentDocumentReadiness = privateData.options.readyState || "loading";

    this._lastFocusedElement = null;

    this._resourceLoader = new PerDocumentResourceLoader(this);

    // Each Document in a browsing context can also have a latest entry. This is the entry for that Document
    // to which the browsing context's session history was most recently traversed. When a Document is created,
    // it initially has no latest entry.
    this._latestEntry = null;

    // https://html.spec.whatwg.org/multipage/dynamic-markup-insertion.html#throw-on-dynamic-markup-insertion-counter
    this._throwOnDynamicMarkupInsertionCounter = 0;
  ***REMOVED***

  _getTheParent(event) {
    if (event.type === "load" || !this._defaultView) {
      return null;
    ***REMOVED***

    return idlUtils.implForWrapper(this._defaultView);
  ***REMOVED***

  get compatMode() {
    return this._parsingMode === "xml" || this.doctype ? "CSS1Compat" : "BackCompat";
  ***REMOVED***
  get charset() {
    return this._encoding;
  ***REMOVED***
  get characterSet() {
    return this._encoding;
  ***REMOVED***
  get inputEncoding() {
    return this._encoding;
  ***REMOVED***
  get doctype() {
    for (const childNode of domSymbolTree.childrenIterator(this)) {
      if (childNode.nodeType === NODE_TYPE.DOCUMENT_TYPE_NODE) {
        return childNode;
      ***REMOVED***
    ***REMOVED***
    return null;
  ***REMOVED***
  get URL() {
    return whatwgURL.serializeURL(this._URL);
  ***REMOVED***
  get documentURI() {
    return whatwgURL.serializeURL(this._URL);
  ***REMOVED***
  get location() {
    return this._defaultView ? this._location : null;
  ***REMOVED***

  // https://dom.spec.whatwg.org/#dom-document-documentelement
  get documentElement() {
    for (const childNode of domSymbolTree.childrenIterator(this)) {
      if (childNode.nodeType === NODE_TYPE.ELEMENT_NODE) {
        return childNode;
      ***REMOVED***
    ***REMOVED***

    return null;
  ***REMOVED***

  get implementation() {
    return this._implementation;
  ***REMOVED***
  set implementation(implementation) {
    this._implementation = implementation;
  ***REMOVED***

  get defaultView() {
    return this._defaultView;
  ***REMOVED***

  get currentScript() {
    return this._currentScript;
  ***REMOVED***

  get readyState() {
    return this._currentDocumentReadiness;
  ***REMOVED***

  set readyState(state) {
    this._currentDocumentReadiness = state;
    fireAnEvent("readystatechange", this);
  ***REMOVED***

  hasFocus() {
    return Boolean(this._lastFocusedElement);
  ***REMOVED***

  _descendantRemoved(parent, child) {
    if (child.tagName === "STYLE") {
      this.styleSheets._remove(child.sheet);
    ***REMOVED***

    super._descendantRemoved(parent, child);
  ***REMOVED***

  write(...args) {
    let text = "";
    for (let i = 0; i < args.length; ++i) {
      text += args[i];
    ***REMOVED***

    if (this._parsingMode === "xml") {
      throw DOMException.create(this._globalObject, [
        "Cannot use document.write on XML documents",
        "InvalidStateError"
      ]);
    ***REMOVED***

    if (this._throwOnDynamicMarkupInsertionCounter > 0) {
      throw DOMException.create(this._globalObject, [
        "Cannot use document.write while a custom element upgrades",
        "InvalidStateError"
      ]);
    ***REMOVED***

    if (this._writeAfterElement) {
      // If called from an script element directly (during the first tick),
      // the new elements are inserted right after that element.
      const tempDiv = this.createElement("div");
      tempDiv.innerHTML = text;

      let child = tempDiv.firstChild;
      let previous = this._writeAfterElement;
      const parent = this._writeAfterElement.parentNode;

      while (child) {
        const node = child;
        child = child.nextSibling;

        node._isMovingDueToDocumentWrite = true; // hack for script execution
        parent.insertBefore(node, previous.nextSibling);
        node._isMovingDueToDocumentWrite = false;

        previous = node;
      ***REMOVED***
    ***REMOVED*** else if (this.readyState === "loading") {
      // During page loading, document.write appends to the current element
      // Find the last child that has been added to the document.
      if (this.lastChild) {
        let node = this;
        while (node.lastChild && node.lastChild.nodeType === NODE_TYPE.ELEMENT_NODE) {
          node = node.lastChild;
        ***REMOVED***
        node.innerHTML = text;
      ***REMOVED*** else {
        clearChildNodes(this);
        parseIntoDocument(text, this);
      ***REMOVED***
    ***REMOVED*** else if (text) {
      clearChildNodes(this);
      parseIntoDocument(text, this);
    ***REMOVED***
  ***REMOVED***

  writeln(...args) {
    this.write(...args, "\n");
  ***REMOVED***

  // This is implemented separately for Document (which has a _ids cache) and DocumentFragment (which does not).
  getElementById(id) {
    if (!this._ids[id]) {
      return null;
    ***REMOVED***

    // Let's find the first element with where it's root is the document.
    const matchElement = this._ids[id].find(candidate => {
      let root = candidate;
      while (domSymbolTree.parent(root)) {
        root = domSymbolTree.parent(root);
      ***REMOVED***

      return root === this;
    ***REMOVED***);

    return matchElement || null;
  ***REMOVED***

  get referrer() {
    return this._referrer || "";
  ***REMOVED***
  get lastModified() {
    return this._lastModified;
  ***REMOVED***
  get images() {
    return this.getElementsByTagName("IMG");
  ***REMOVED***
  get embeds() {
    return this.getElementsByTagName("EMBED");
  ***REMOVED***
  get plugins() {
    return this.embeds;
  ***REMOVED***
  get links() {
    return HTMLCollection.createImpl(this._globalObject, [], {
      element: this,
      query: () => domSymbolTree.treeToArray(this, {
        filter: node => (node._localName === "a" || node._localName === "area") &&
                        node.hasAttributeNS(null, "href") &&
                        node._namespaceURI === HTML_NS
      ***REMOVED***)
    ***REMOVED***);
  ***REMOVED***
  get forms() {
    return this.getElementsByTagName("FORM");
  ***REMOVED***
  get scripts() {
    return this.getElementsByTagName("SCRIPT");
  ***REMOVED***
  get anchors() {
    return HTMLCollection.createImpl(this._globalObject, [], {
      element: this,
      query: () => domSymbolTree.treeToArray(this, {
        filter: node => node._localName === "a" &&
                        node.hasAttributeNS(null, "name") &&
                        node._namespaceURI === HTML_NS
      ***REMOVED***)
    ***REMOVED***);
  ***REMOVED***

  // The applets attribute must return an
  // HTMLCollection rooted at the Document node,
  // whose filter matches nothing.
  // (It exists for historical reasons.)
  get applets() {
    return HTMLCollection.createImpl(this._globalObject, [], {
      element: this,
      query: () => []
    ***REMOVED***);
  ***REMOVED***

  open() {
    let child = domSymbolTree.firstChild(this);
    while (child) {
      this.removeChild(child);
      child = domSymbolTree.firstChild(this);
    ***REMOVED***
    this._modified();
    return this;
  ***REMOVED***
  close(noQueue) {
    // In some cases like when creating an empty iframe, I want to emit the
    // events right away to avoid problems if later I asign the property src.
    if (noQueue) {
      this.readyState = "complete";

      fireAnEvent("DOMContentLoaded", this, undefined, { bubbles: true ***REMOVED***);
      fireAnEvent("load", this);

      return;
    ***REMOVED***
    this._queue.resume();

    const dummyPromise = Promise.resolve();

    const onDOMContentLoad = () => {
      const doc = this;
      function dispatchEvent() {
        // https://html.spec.whatwg.org/#the-end
        doc.readyState = "interactive";
        fireAnEvent("DOMContentLoaded", doc, undefined, { bubbles: true ***REMOVED***);
      ***REMOVED***

      return new Promise(resolve => {
        if (!this._deferQueue.tail) {
          dispatchEvent();
          resolve();
          return;
        ***REMOVED***

        this._deferQueue.setListener(() => {
          dispatchEvent();
          resolve();
        ***REMOVED***);

        this._deferQueue.resume();
      ***REMOVED***);
    ***REMOVED***;

    const onLoad = () => {
      const doc = this;
      function dispatchEvent() {
        doc.readyState = "complete";
        fireAnEvent("load", doc);
      ***REMOVED***

      return new Promise(resolve => {
        if (this._asyncQueue.count() === 0) {
          dispatchEvent();
          resolve();
          return;
        ***REMOVED***

        this._asyncQueue.setListener(() => {
          dispatchEvent();
          resolve();
        ***REMOVED***);
      ***REMOVED***);
    ***REMOVED***;

    this._queue.push(dummyPromise, onDOMContentLoad, null);
    // Set the readyState to 'complete' once all resources are loaded.
    // As a side-effect the document's load-event will be dispatched.
    this._queue.push(dummyPromise, onLoad, null, true);
  ***REMOVED***

  getElementsByName(elementName) {
    return NodeList.createImpl(this._globalObject, [], {
      element: this,
      query: () => domSymbolTree.treeToArray(this, {
        filter: node => node.getAttributeNS && node.getAttributeNS(null, "name") === elementName
      ***REMOVED***)
    ***REMOVED***);
  ***REMOVED***

  get title() {
    const { documentElement ***REMOVED*** = this;
    let value = "";

    if (documentElement && documentElement._localName === "svg") {
      const svgTitleElement = firstChildWithLocalName(documentElement, "title", SVG_NS);

      if (svgTitleElement) {
        value = childTextContent(svgTitleElement);
      ***REMOVED***
    ***REMOVED*** else {
      const titleElement = firstDescendantWithLocalName(this, "title");

      if (titleElement) {
        value = childTextContent(titleElement);
      ***REMOVED***
    ***REMOVED***

    value = stripAndCollapseASCIIWhitespace(value);

    return value;
  ***REMOVED***

  set title(value) {
    const { documentElement ***REMOVED*** = this;
    let element;

    if (documentElement && documentElement._localName === "svg") {
      element = firstChildWithLocalName(documentElement, "title", SVG_NS);

      if (!element) {
        element = this.createElementNS(SVG_NS, "title");

        this._insert(element, documentElement.firstChild);
      ***REMOVED***

      element.textContent = value;
    ***REMOVED*** else if (documentElement && documentElement._namespaceURI === HTML_NS) {
      const titleElement = firstDescendantWithLocalName(this, "title");
      const headElement = this.head;

      if (titleElement === null && headElement === null) {
        return;
      ***REMOVED***

      if (titleElement !== null) {
        element = titleElement;
      ***REMOVED*** else {
        element = this.createElement("title");
        headElement._append(element);
      ***REMOVED***

      element.textContent = value;
    ***REMOVED***
  ***REMOVED***

  get dir() {
    return this.documentElement ? this.documentElement.dir : "";
  ***REMOVED***
  set dir(value) {
    if (this.documentElement) {
      this.documentElement.dir = value;
    ***REMOVED***
  ***REMOVED***

  get head() {
    return this.documentElement ? firstChildWithLocalName(this.documentElement, "head") : null;
  ***REMOVED***

  get body() {
    const { documentElement ***REMOVED*** = this;
    if (!documentElement || documentElement._localName !== "html" ||
        documentElement._namespaceURI !== HTML_NS) {
      return null;
    ***REMOVED***

    return firstChildWithLocalNames(this.documentElement, new Set(["body", "frameset"]));
  ***REMOVED***

  set body(value) {
    if (value === null ||
        value._namespaceURI !== HTML_NS ||
        (value._localName !== "body" && value._localName !== "frameset")) {
      throw DOMException.create(this._globalObject, [
        "Cannot set the body to null or a non-body/frameset element",
        "HierarchyRequestError"
      ]);
    ***REMOVED***

    const bodyElement = this.body;
    if (value === bodyElement) {
      return;
    ***REMOVED***

    if (bodyElement !== null) {
      bodyElement.parentNode._replace(value, bodyElement);
      return;
    ***REMOVED***

    const { documentElement ***REMOVED*** = this;
    if (documentElement === null) {
      throw DOMException.create(this._globalObject, [
        "Cannot set the body when there is no document element",
        "HierarchyRequestError"
      ]);
    ***REMOVED***

    documentElement._append(value);
  ***REMOVED***

  _runPreRemovingSteps(oldNode) {
    // https://html.spec.whatwg.org/#focus-fixup-rule
    if (oldNode === this.activeElement) {
      this._lastFocusedElement = this.body;
    ***REMOVED***
    for (const activeNodeIterator of this._workingNodeIterators) {
      activeNodeIterator._preRemovingSteps(oldNode);
    ***REMOVED***
  ***REMOVED***

  createEvent(type) {
    const typeLower = type.toLowerCase();
    const eventWrapper = eventInterfaceTable[typeLower] || null;

    if (!eventWrapper) {
      throw DOMException.create(this._globalObject, [
        "The provided event type (\"" + type + "\") is invalid",
        "NotSupportedError"
      ]);
    ***REMOVED***

    const impl = eventWrapper.createImpl(this._globalObject, [""]);
    impl._initializedFlag = false;
    return impl;
  ***REMOVED***

  createRange() {
    return Range.createImpl(this._globalObject, [], {
      start: { node: this, offset: 0 ***REMOVED***,
      end: { node: this, offset: 0 ***REMOVED***
    ***REMOVED***);
  ***REMOVED***

  createProcessingInstruction(target, data) {
    validateName(this._globalObject, target);

    if (data.includes("?>")) {
      throw DOMException.create(this._globalObject, [
        "Processing instruction data cannot contain the string \"?>\"",
        "InvalidCharacterError"
      ]);
    ***REMOVED***

    return ProcessingInstruction.createImpl(this._globalObject, [], {
      ownerDocument: this,
      target,
      data
    ***REMOVED***);
  ***REMOVED***

  // https://dom.spec.whatwg.org/#dom-document-createcdatasection
  createCDATASection(data) {
    if (this._parsingMode === "html") {
      throw DOMException.create(this._globalObject, [
        "Cannot create CDATA sections in HTML documents",
        "NotSupportedError"
      ]);
    ***REMOVED***

    if (data.includes("]]>")) {
      throw DOMException.create(this._globalObject, [
        "CDATA section data cannot contain the string \"]]>\"",
        "InvalidCharacterError"
      ]);
    ***REMOVED***

    return CDATASection.createImpl(this._globalObject, [], {
      ownerDocument: this,
      data
    ***REMOVED***);
  ***REMOVED***

  createTextNode(data) {
    return Text.createImpl(this._globalObject, [], {
      ownerDocument: this,
      data
    ***REMOVED***);
  ***REMOVED***

  createComment(data) {
    return Comment.createImpl(this._globalObject, [], {
      ownerDocument: this,
      data
    ***REMOVED***);
  ***REMOVED***

  // https://dom.spec.whatwg.org/#dom-document-createelement
  createElement(localName, options) {
    validateName(this._globalObject, localName);

    if (this._parsingMode === "html") {
      localName = asciiLowercase(localName);
    ***REMOVED***

    let isValue = null;
    if (options && options.is !== undefined) {
      isValue = options.is;
    ***REMOVED***

    const namespace = this._parsingMode === "html" || this.contentType === "application/xhtml+xml" ? HTML_NS : null;

    return createElement(this, localName, namespace, null, isValue, true);
  ***REMOVED***

  // https://dom.spec.whatwg.org/#dom-document-createelementns
  createElementNS(namespace, qualifiedName, options) {
    return internalCreateElementNSSteps(this, namespace, qualifiedName, options);
  ***REMOVED***

  createDocumentFragment() {
    return DocumentFragment.createImpl(this._globalObject, [], { ownerDocument: this ***REMOVED***);
  ***REMOVED***

  createAttribute(localName) {
    validateName(this._globalObject, localName);

    if (this._parsingMode === "html") {
      localName = asciiLowercase(localName);
    ***REMOVED***

    return this._createAttribute({ localName ***REMOVED***);
  ***REMOVED***

  createAttributeNS(namespace, name) {
    if (namespace === undefined) {
      namespace = null;
    ***REMOVED***
    namespace = namespace !== null ? String(namespace) : namespace;

    const extracted = validateAndExtract(this._globalObject, namespace, name);
    return this._createAttribute({
      namespace: extracted.namespace,
      namespacePrefix: extracted.prefix,
      localName: extracted.localName
    ***REMOVED***);
  ***REMOVED***

  // Using this helper function rather than directly calling generatedAttr.createImpl may be preferred in some files,
  // to avoid introducing a potentially cyclic dependency on generated/Attr.js.
  _createAttribute({
    localName,
    value,
    namespace,
    namespacePrefix
  ***REMOVED***) {
    return generatedAttr.createImpl(this._globalObject, [], {
      localName,
      value,
      namespace,
      namespacePrefix,
      ownerDocument: this
    ***REMOVED***);
  ***REMOVED***

  createTreeWalker(root, whatToShow, filter) {
    return TreeWalker.createImpl(this._globalObject, [], { root, whatToShow, filter ***REMOVED***);
  ***REMOVED***

  createNodeIterator(root, whatToShow, filter) {
    const nodeIterator = NodeIterator.createImpl(this._globalObject, [], { root, whatToShow, filter ***REMOVED***);

    if (hasWeakRefs) {
      this._workingNodeIterators.add(nodeIterator);
    ***REMOVED*** else {
      this._workingNodeIterators.push(nodeIterator);
      while (this._workingNodeIterators.length > 10) {
        const toInactivate = this._workingNodeIterators.shift();
        toInactivate._working = false;
      ***REMOVED***
    ***REMOVED***

    return nodeIterator;
  ***REMOVED***

  importNode(node, deep) {
    if (node.nodeType === NODE_TYPE.DOCUMENT_NODE) {
      throw DOMException.create(this._globalObject, [
        "Cannot import a document node",
        "NotSupportedError"
      ]);
    ***REMOVED*** else if (ShadowRoot.isImpl(node)) {
      throw DOMException.create(this._globalObject, [
        "Cannot adopt a shadow root",
        "NotSupportedError"
      ]);
    ***REMOVED***

    return clone(node, this, deep);
  ***REMOVED***

  // https://dom.spec.whatwg.org/#dom-document-adoptnode
  adoptNode(node) {
    if (node.nodeType === NODE_TYPE.DOCUMENT_NODE) {
      throw DOMException.create(this._globalObject, [
        "Cannot adopt a document node",
        "NotSupportedError"
      ]);
    ***REMOVED*** else if (ShadowRoot.isImpl(node)) {
      throw DOMException.create(this._globalObject, [
        "Cannot adopt a shadow root",
        "HierarchyRequestError"
      ]);
    ***REMOVED***

    this._adoptNode(node);

    return node;
  ***REMOVED***

  // https://dom.spec.whatwg.org/#concept-node-adopt
  _adoptNode(node) {
    const newDocument = this;
    const oldDocument = node._ownerDocument;

    const parent = domSymbolTree.parent(node);
    if (parent) {
      parent._remove(node);
    ***REMOVED***

    if (oldDocument !== newDocument) {
      for (const inclusiveDescendant of shadowIncludingInclusiveDescendantsIterator(node)) {
        inclusiveDescendant._ownerDocument = newDocument;
      ***REMOVED***

      for (const inclusiveDescendant of shadowIncludingInclusiveDescendantsIterator(node)) {
        if (inclusiveDescendant._ceState === "custom") {
          enqueueCECallbackReaction(inclusiveDescendant, "adoptedCallback", [
            idlUtils.wrapperForImpl(oldDocument),
            idlUtils.wrapperForImpl(newDocument)
          ]);
        ***REMOVED***
      ***REMOVED***

      for (const inclusiveDescendant of shadowIncludingInclusiveDescendantsIterator(node)) {
        if (inclusiveDescendant._adoptingSteps) {
          inclusiveDescendant._adoptingSteps(oldDocument);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  get cookie() {
    return this._cookieJar.getCookieStringSync(this.URL, { http: false ***REMOVED***);
  ***REMOVED***
  set cookie(cookieStr) {
    cookieStr = String(cookieStr);
    this._cookieJar.setCookieSync(cookieStr, this.URL, {
      http: false,
      ignoreError: true
    ***REMOVED***);
  ***REMOVED***

  // The clear(), captureEvents(), and releaseEvents() methods must do nothing
  clear() {***REMOVED***

  captureEvents() {***REMOVED***

  releaseEvents() {***REMOVED***

  get styleSheets() {
    if (!this._styleSheets) {
      this._styleSheets = StyleSheetList.createImpl(this._globalObject);
    ***REMOVED***

    // TODO: each style and link element should register its sheet on creation
    // and remove it on removal.
    return this._styleSheets;
  ***REMOVED***

  get hidden() {
    if (this._defaultView && this._defaultView._pretendToBeVisual) {
      return false;
    ***REMOVED***

    return true;
  ***REMOVED***

  get visibilityState() {
    if (this._defaultView && this._defaultView._pretendToBeVisual) {
      return "visible";
    ***REMOVED***

    return "prerender";
  ***REMOVED***

  // https://w3c.github.io/selection-api/#extensions-to-document-interface
  getSelection() {
    return this._defaultView ? this._defaultView._selection : null;
  ***REMOVED***

  // Needed to ensure that the resulting document has the correct prototype chain:
  // https://dom.spec.whatwg.org/#concept-node-clone says "that implements the same interfaces as node".
  _cloneDocument() {
    const copy = documents.createImpl(
      this._globalObject,
      {
        contentType: this.contentType,
        encoding: this._encoding,
        parsingMode: this._parsingMode
      ***REMOVED***
    );

    copy._URL = this._URL;
    copy._origin = this._origin;
    return copy;
  ***REMOVED***
***REMOVED***

eventAccessors.createEventAccessor(DocumentImpl.prototype, "readystatechange");
mixin(DocumentImpl.prototype, DocumentOrShadowRootImpl.prototype);
mixin(DocumentImpl.prototype, GlobalEventHandlersImpl.prototype);
mixin(DocumentImpl.prototype, NonElementParentNodeImpl.prototype);
mixin(DocumentImpl.prototype, ParentNodeImpl.prototype);

DocumentImpl.prototype.getElementsByTagName = memoizeQuery(function (qualifiedName) {
  return listOfElementsWithQualifiedName(qualifiedName, this);
***REMOVED***);

DocumentImpl.prototype.getElementsByTagNameNS = memoizeQuery(function (namespace, localName) {
  return listOfElementsWithNamespaceAndLocalName(namespace, localName, this);
***REMOVED***);

DocumentImpl.prototype.getElementsByClassName = memoizeQuery(function getElementsByClassName(classNames) {
  return listOfElementsWithClassNames(classNames, this);
***REMOVED***);

module.exports = {
  implementation: DocumentImpl
***REMOVED***;
