// jsdom/lib/jsdom/living/nodes/DOMTokenList-impl.js
"use strict";

const DOMException = require("domexception/webidl2js-wrapper");
const OrderedSet = require("../helpers/ordered-set.js");
const { asciiLowercase ***REMOVED*** = require("../helpers/strings.js");
const idlUtils = require("../generated/utils.js");

const { getAttributeValue, setAttributeValue, hasAttributeByName ***REMOVED*** = require("../attributes.js");

function validateTokens(globalObject, ...tokens) {
  for (const token of tokens) {
    if (token === "") {
      throw DOMException.create(globalObject, ["The token provided must not be empty.", "SyntaxError"]);
    ***REMOVED***
  ***REMOVED***
  for (const token of tokens) {
    if (/[\t\n\f\r ]/.test(token)) {
      throw DOMException.create(globalObject, [
        "The token provided contains HTML space characters, which are not valid in tokens.",
        "InvalidCharacterError"
      ]);
    ***REMOVED***
  ***REMOVED***
***REMOVED***

// https://dom.spec.whatwg.org/#domtokenlist
class DOMTokenListImpl {
  constructor(globalObject, args, privateData) {
    this._globalObject = globalObject;

    // _syncWithElement() must always be called before any _tokenSet access.
    this._tokenSet = new OrderedSet();
    this._element = privateData.element;
    this._attributeLocalName = privateData.attributeLocalName;
    this._supportedTokens = privateData.supportedTokens;

    // Needs synchronization with element if token set is to be accessed.
    this._dirty = true;
  ***REMOVED***

  attrModified() {
    this._dirty = true;
  ***REMOVED***

  _syncWithElement() {
    if (!this._dirty) {
      return;
    ***REMOVED***

    const val = getAttributeValue(this._element, this._attributeLocalName);
    if (val === null) {
      this._tokenSet.empty();
    ***REMOVED*** else {
      this._tokenSet = OrderedSet.parse(val);
    ***REMOVED***

    this._dirty = false;
  ***REMOVED***

  _validationSteps(token) {
    if (!this._supportedTokens) {
      throw new TypeError(`${this._attributeLocalName***REMOVED*** attribute has no supported tokens`);
    ***REMOVED***
    const lowerToken = asciiLowercase(token);
    return this._supportedTokens.has(lowerToken);
  ***REMOVED***

  _updateSteps() {
    if (!hasAttributeByName(this._element, this._attributeLocalName) && this._tokenSet.isEmpty()) {
      return;
    ***REMOVED***
    setAttributeValue(this._element, this._attributeLocalName, this._tokenSet.serialize());
  ***REMOVED***

  _serializeSteps() {
    return getAttributeValue(this._element, this._attributeLocalName);
  ***REMOVED***

  // Used by other parts of jsdom
  get tokenSet() {
    this._syncWithElement();
    return this._tokenSet;
  ***REMOVED***

  get length() {
    this._syncWithElement();
    return this._tokenSet.size;
  ***REMOVED***

  get [idlUtils.supportedPropertyIndices]() {
    this._syncWithElement();
    return this._tokenSet.keys();
  ***REMOVED***

  item(index) {
    this._syncWithElement();
    if (index >= this._tokenSet.size) {
      return null;
    ***REMOVED***
    return this._tokenSet.get(index);
  ***REMOVED***

  contains(token) {
    this._syncWithElement();
    return this._tokenSet.contains(token);
  ***REMOVED***

  add(...tokens) {
    for (const token of tokens) {
      validateTokens(this._globalObject, token);
    ***REMOVED***
    this._syncWithElement();
    for (const token of tokens) {
      this._tokenSet.append(token);
    ***REMOVED***
    this._updateSteps();
  ***REMOVED***

  remove(...tokens) {
    for (const token of tokens) {
      validateTokens(this._globalObject, token);
    ***REMOVED***
    this._syncWithElement();
    this._tokenSet.remove(...tokens);
    this._updateSteps();
  ***REMOVED***

  toggle(token, force = undefined) {
    validateTokens(this._globalObject, token);
    this._syncWithElement();
    if (this._tokenSet.contains(token)) {
      if (force === undefined || force === false) {
        this._tokenSet.remove(token);
        this._updateSteps();
        return false;
      ***REMOVED***
      return true;
    ***REMOVED***
    if (force === undefined || force === true) {
      this._tokenSet.append(token);
      this._updateSteps();
      return true;
    ***REMOVED***
    return false;
  ***REMOVED***

  replace(token, newToken) {
    validateTokens(this._globalObject, token, newToken);
    this._syncWithElement();
    if (!this._tokenSet.contains(token)) {
      return false;
    ***REMOVED***
    this._tokenSet.replace(token, newToken);
    this._updateSteps();
    return true;
  ***REMOVED***

  supports(token) {
    return this._validationSteps(token);
  ***REMOVED***

  get value() {
    return this._serializeSteps();
  ***REMOVED***

  set value(V) {
    setAttributeValue(this._element, this._attributeLocalName, V);
  ***REMOVED***
***REMOVED***

exports.implementation = DOMTokenListImpl;
