// jsdom/lib/jsdom/living/nodes/HTMLTableElement-impl.js
"use strict";
const DOMException = require("domexception/webidl2js-wrapper");
const HTMLElementImpl = require("./HTMLElement-impl").implementation;
const { HTML_NS ***REMOVED*** = require("../helpers/namespaces");
const { domSymbolTree ***REMOVED*** = require("../helpers/internal-constants");
const { firstChildWithLocalName, childrenByLocalName ***REMOVED*** = require("../helpers/traversal");
const HTMLCollection = require("../generated/HTMLCollection");
const NODE_TYPE = require("../node-type");

function tHeadInsertionPoint(table) {
  const iterator = domSymbolTree.childrenIterator(table);
  for (const child of iterator) {
    if (child.nodeType !== NODE_TYPE.ELEMENT_NODE) {
      continue;
    ***REMOVED***

    if (child._namespaceURI !== HTML_NS || (child._localName !== "caption" && child._localName !== "colgroup")) {
      return child;
    ***REMOVED***
  ***REMOVED***

  return null;
***REMOVED***

class HTMLTableElementImpl extends HTMLElementImpl {
  get caption() {
    return firstChildWithLocalName(this, "caption");
  ***REMOVED***

  set caption(value) {
    const currentCaption = this.caption;
    if (currentCaption !== null) {
      this.removeChild(currentCaption);
    ***REMOVED***

    if (value !== null) {
      const insertionPoint = this.firstChild;
      this.insertBefore(value, insertionPoint);
    ***REMOVED***
  ***REMOVED***

  get tHead() {
    return firstChildWithLocalName(this, "thead");
  ***REMOVED***

  set tHead(value) {
    if (value !== null && value._localName !== "thead") {
      throw DOMException.create(this._globalObject, [
        "Cannot set a non-thead element as a table header",
        "HierarchyRequestError"
      ]);
    ***REMOVED***

    const currentHead = this.tHead;
    if (currentHead !== null) {
      this.removeChild(currentHead);
    ***REMOVED***

    if (value !== null) {
      const insertionPoint = tHeadInsertionPoint(this);
      this.insertBefore(value, insertionPoint);
    ***REMOVED***
  ***REMOVED***

  get tFoot() {
    return firstChildWithLocalName(this, "tfoot");
  ***REMOVED***

  set tFoot(value) {
    if (value !== null && value._localName !== "tfoot") {
      throw DOMException.create(this._globalObject, [
        "Cannot set a non-tfoot element as a table footer",
        "HierarchyRequestError"
      ]);
    ***REMOVED***

    const currentFoot = this.tFoot;
    if (currentFoot !== null) {
      this.removeChild(currentFoot);
    ***REMOVED***

    if (value !== null) {
      this.appendChild(value);
    ***REMOVED***
  ***REMOVED***

  get rows() {
    if (!this._rows) {
      this._rows = HTMLCollection.createImpl(this._globalObject, [], {
        element: this,
        query: () => {
          const headerRows = [];
          const bodyRows = [];
          const footerRows = [];

          const iterator = domSymbolTree.childrenIterator(this);
          for (const child of iterator) {
            if (child.nodeType !== NODE_TYPE.ELEMENT_NODE || child._namespaceURI !== HTML_NS) {
              continue;
            ***REMOVED***

            if (child._localName === "thead") {
              headerRows.push(...childrenByLocalName(child, "tr"));
            ***REMOVED*** else if (child._localName === "tbody") {
              bodyRows.push(...childrenByLocalName(child, "tr"));
            ***REMOVED*** else if (child._localName === "tfoot") {
              footerRows.push(...childrenByLocalName(child, "tr"));
            ***REMOVED*** else if (child._localName === "tr") {
              bodyRows.push(child);
            ***REMOVED***
          ***REMOVED***

          return [...headerRows, ...bodyRows, ...footerRows];
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***
    return this._rows;
  ***REMOVED***

  get tBodies() {
    if (!this._tBodies) {
      this._tBodies = HTMLCollection.createImpl(this._globalObject, [], {
        element: this,
        query: () => childrenByLocalName(this, "tbody")
      ***REMOVED***);
    ***REMOVED***
    return this._tBodies;
  ***REMOVED***

  createTBody() {
    const el = this._ownerDocument.createElement("TBODY");

    const tbodies = childrenByLocalName(this, "tbody");
    const insertionPoint = tbodies[tbodies.length - 1];

    if (insertionPoint) {
      this.insertBefore(el, insertionPoint.nextSibling);
    ***REMOVED*** else {
      this.appendChild(el);
    ***REMOVED***
    return el;
  ***REMOVED***

  createTHead() {
    let el = this.tHead;
    if (!el) {
      el = this.tHead = this._ownerDocument.createElement("THEAD");
    ***REMOVED***
    return el;
  ***REMOVED***

  deleteTHead() {
    this.tHead = null;
  ***REMOVED***

  createTFoot() {
    let el = this.tFoot;
    if (!el) {
      el = this.tFoot = this._ownerDocument.createElement("TFOOT");
    ***REMOVED***
    return el;
  ***REMOVED***

  deleteTFoot() {
    this.tFoot = null;
  ***REMOVED***

  createCaption() {
    let el = this.caption;
    if (!el) {
      el = this.caption = this._ownerDocument.createElement("CAPTION");
    ***REMOVED***
    return el;
  ***REMOVED***

  deleteCaption() {
    const c = this.caption;
    if (c) {
      c.parentNode.removeChild(c);
    ***REMOVED***
  ***REMOVED***

  insertRow(index) {
    if (index < -1 || index > this.rows.length) {
      throw DOMException.create(this._globalObject, [
        "Cannot insert a row at an index that is less than -1 or greater than the number of existing rows",
        "IndexSizeError"
      ]);
    ***REMOVED***

    const tr = this._ownerDocument.createElement("tr");

    if (this.rows.length === 0 && this.tBodies.length === 0) {
      const tBody = this._ownerDocument.createElement("tbody");
      tBody.appendChild(tr);
      this.appendChild(tBody);
    ***REMOVED*** else if (this.rows.length === 0) {
      const tBody = this.tBodies.item(this.tBodies.length - 1);
      tBody.appendChild(tr);
    ***REMOVED*** else if (index === -1 || index === this.rows.length) {
      const tSection = this.rows.item(this.rows.length - 1).parentNode;
      tSection.appendChild(tr);
    ***REMOVED*** else {
      const beforeTR = this.rows.item(index);
      const tSection = beforeTR.parentNode;
      tSection.insertBefore(tr, beforeTR);
    ***REMOVED***

    return tr;
  ***REMOVED***

  deleteRow(index) {
    const rowLength = this.rows.length;
    if (index < -1 || index >= rowLength) {
      throw DOMException.create(this._globalObject, [
        `Cannot delete a row at index ${index***REMOVED***, where no row exists`,
        "IndexSizeError"
      ]);
    ***REMOVED***

    if (index === -1) {
      if (rowLength === 0) {
        return;
      ***REMOVED***

      index = rowLength - 1;
    ***REMOVED***

    const tr = this.rows.item(index);
    tr.parentNode.removeChild(tr);
  ***REMOVED***
***REMOVED***

module.exports = {
  implementation: HTMLTableElementImpl
***REMOVED***;
