// jsdom/lib/jsdom/living/nodes/HTMLTableRowElement-impl.js
"use strict";

const DOMException = require("domexception/webidl2js-wrapper");
const HTMLElementImpl = require("./HTMLElement-impl").implementation;
const HTMLCollection = require("../generated/HTMLCollection");
const { HTML_NS ***REMOVED*** = require("../helpers/namespaces");
const { childrenByLocalNames ***REMOVED*** = require("../helpers/traversal");
const { domSymbolTree ***REMOVED*** = require("../helpers/internal-constants");

const cellLocalNames = new Set(["td", "th"]);

class HTMLTableRowElementImpl extends HTMLElementImpl {
  get cells() {
    if (!this._cells) {
      this._cells = HTMLCollection.createImpl(this._globalObject, [], {
        element: this,
        query: () => childrenByLocalNames(this, cellLocalNames)
      ***REMOVED***);
    ***REMOVED***
    return this._cells;
  ***REMOVED***

  get rowIndex() {
    const parent = this.parentElement;
    if (parent === null || parent.namespaceURI !== HTML_NS) {
      return -1;
    ***REMOVED***

    let tableElement = parent;
    if (parent.localName === "thead" || parent.localName === "tbody" || parent.localName === "tfoot") {
      tableElement = parent.parentElement;
    ***REMOVED***
    if (tableElement === null || tableElement.namespaceURI !== HTML_NS || tableElement.localName !== "table") {
      return -1;
    ***REMOVED***

    return tableElement.rows.indexOf(this);
  ***REMOVED***

  get sectionRowIndex() {
    const parent = domSymbolTree.parent(this);
    if (parent === null) {
      return -1;
    ***REMOVED***

    const { rows ***REMOVED*** = parent;
    if (!rows) {
      return -1;
    ***REMOVED***

    return rows.indexOf(this);
  ***REMOVED***

  insertCell(index) {
    const td = this._ownerDocument.createElement("TD");
    const { cells ***REMOVED*** = this;
    if (index < -1 || index > cells.length) {
      throw DOMException.create(this._globalObject, ["The index is not in the allowed range.", "IndexSizeError"]);
    ***REMOVED***
    if (index === -1 || index === cells.length) {
      this._append(td);
    ***REMOVED*** else {
      const ref = cells.item(index);
      this._insert(td, ref);
    ***REMOVED***
    return td;
  ***REMOVED***

  deleteCell(index) {
    const { cells ***REMOVED*** = this;
    if (index < -1 || index >= cells.length) {
      throw DOMException.create(this._globalObject, ["The index is not in the allowed range.", "IndexSizeError"]);
    ***REMOVED***
    if (index === -1) {
      if (cells.length === 0) {
        return;
      ***REMOVED***

      index = cells.length - 1;
    ***REMOVED***
    const td = cells.item(index);
    this._remove(td);
  ***REMOVED***
***REMOVED***

module.exports = {
  implementation: HTMLTableRowElementImpl
***REMOVED***;
