// jsdom/lib/jsdom/living/nodes/HTMLCollection-impl.js
"use strict";

const idlUtils = require("../generated/utils.js");
const { HTML_NS ***REMOVED*** = require("../helpers/namespaces");

exports.implementation = class HTMLCollectionImpl {
  constructor(globalObject, args, privateData) {
    this._list = [];
    this._version = -1;
    this._element = privateData.element;
    this._query = privateData.query;

    this._globalObject = globalObject;

    this._update();
  ***REMOVED***
  get length() {
    this._update();
    return this._list.length;
  ***REMOVED***
  item(index) {
    this._update();
    return this._list[index] || null;
  ***REMOVED***
  namedItem(key) {
    if (key === "") {
      return null;
    ***REMOVED***
    this._update();
    for (const element of this._list) {
      if (element.getAttributeNS(null, "id") === key) {
        return element;
      ***REMOVED***
      if (element._namespaceURI === HTML_NS) {
        const name = element.getAttributeNS(null, "name");
        if (name === key) {
          return element;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
    return null;
  ***REMOVED***
  _update() {
    if (this._version < this._element._version) {
      const snapshot = this._query();
      for (let i = 0; i < snapshot.length; i++) {
        this._list[i] = snapshot[i];
      ***REMOVED***
      this._list.length = snapshot.length;
      this._version = this._element._version;
    ***REMOVED***
  ***REMOVED***
  get [idlUtils.supportedPropertyIndices]() {
    this._update();
    return this._list.keys();
  ***REMOVED***
  get [idlUtils.supportedPropertyNames]() {
    this._update();
    const result = new Set();
    for (const element of this._list) {
      const id = element.getAttributeNS(null, "id");
      if (id) {
        result.add(id);
      ***REMOVED***
      if (element._namespaceURI === HTML_NS) {
        const name = element.getAttributeNS(null, "name");
        if (name) {
          result.add(name);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
    return result;
  ***REMOVED***

  // Inherit some useful functions from Array.
  [Symbol.iterator]() {
    this._update();
    return this._list[Symbol.iterator]();
  ***REMOVED***
  entries() {
    this._update();
    return this._list.entries();
  ***REMOVED***
  filter(...args) {
    this._update();
    return this._list.filter(...args);
  ***REMOVED***
  map(...args) {
    this._update();
    return this._list.map(...args);
  ***REMOVED***
  indexOf(...args) {
    this._update();
    return this._list.indexOf(...args);
  ***REMOVED***
***REMOVED***;
