// jsdom/lib/jsdom/living/nodes/HTMLMediaElement-impl.js
"use strict";
const DOMException = require("domexception/webidl2js-wrapper");
const HTMLElementImpl = require("./HTMLElement-impl").implementation;
const notImplemented = require("../../browser/not-implemented");
const { fireAnEvent ***REMOVED*** = require("../helpers/events");

function getTimeRangeDummy() {
  return {
    length: 0,
    start() {
      return 0;
    ***REMOVED***,
    end() {
      return 0;
    ***REMOVED***
  ***REMOVED***;
***REMOVED***

class HTMLMediaElementImpl extends HTMLElementImpl {
  constructor(globalObject, args, privateData) {
    super(globalObject, args, privateData);

    this._muted = false;
    this._volume = 1.0;
    this.readyState = 0;
    this.networkState = 0;
    this.currentTime = 0;
    this.currentSrc = "";
    this.buffered = getTimeRangeDummy();
    this.seeking = false;
    this.duration = NaN;
    this.paused = true;
    this.played = getTimeRangeDummy();
    this.seekable = getTimeRangeDummy();
    this.ended = false;
    this.audioTracks = [];
    this.videoTracks = [];
    this.textTracks = [];
  ***REMOVED***

  get defaultPlaybackRate() {
    if (this._defaultPlaybackRate === undefined) {
      return 1.0;
    ***REMOVED***
    return this._defaultPlaybackRate;
  ***REMOVED***

  set defaultPlaybackRate(v) {
    if (v === 0.0) {
      throw DOMException.create(this._globalObject, ["The operation is not supported.", "NotSupportedError"]);
    ***REMOVED***
    if (this._defaultPlaybackRate !== v) {
      this._defaultPlaybackRate = v;
      this._dispatchRateChange();
    ***REMOVED***
  ***REMOVED***

  get playbackRate() {
    if (this._playbackRate === undefined) {
      return 1.0;
    ***REMOVED***
    return this._playbackRate;
  ***REMOVED***

  set playbackRate(v) {
    if (v !== this._playbackRate) {
      this._playbackRate = v;
      this._dispatchRateChange();
    ***REMOVED***
  ***REMOVED***

  get muted() {
    return this._muted;
  ***REMOVED***

  set muted(v) {
    if (v !== this._muted) {
      this._muted = v;
      this._dispatchVolumeChange();
    ***REMOVED***
  ***REMOVED***

  get defaultMuted() {
    return this.getAttributeNS(null, "muted") !== null;
  ***REMOVED***

  set defaultMuted(v) {
    if (v) {
      this.setAttributeNS(null, "muted", v);
    ***REMOVED*** else {
      this.removeAttributeNS(null, "muted");
    ***REMOVED***
  ***REMOVED***

  get volume() {
    return this._volume;
  ***REMOVED***

  set volume(v) {
    if (v < 0 || v > 1) {
      throw DOMException.create(this._globalObject, ["The index is not in the allowed range.", "IndexSizeError"]);
    ***REMOVED***
    if (this._volume !== v) {
      this._volume = v;
      this._dispatchVolumeChange();
    ***REMOVED***
  ***REMOVED***

  // Not (yet) implemented according to spec
  // Should return sane default values
  load() {
    notImplemented("HTMLMediaElement.prototype.load", this._ownerDocument._defaultView);
  ***REMOVED***
  canPlayType() {
    return "";
  ***REMOVED***
  play() {
    notImplemented("HTMLMediaElement.prototype.play", this._ownerDocument._defaultView);
  ***REMOVED***
  pause() {
    notImplemented("HTMLMediaElement.prototype.pause", this._ownerDocument._defaultView);
  ***REMOVED***
  addTextTrack() {
    notImplemented("HTMLMediaElement.prototype.addTextTrack", this._ownerDocument._defaultView);
  ***REMOVED***

  _dispatchRateChange() {
    fireAnEvent("ratechange", this);
  ***REMOVED***

  _dispatchVolumeChange() {
    fireAnEvent("volumechange", this);
  ***REMOVED***
***REMOVED***

module.exports = {
  implementation: HTMLMediaElementImpl
***REMOVED***;
