// jsdom/lib/jsdom/living/nodes/HTMLOptionElement-impl.js
"use strict";

const HTMLElementImpl = require("./HTMLElement-impl").implementation;
const NODE_TYPE = require("../node-type");
const { stripAndCollapseASCIIWhitespace ***REMOVED*** = require("../helpers/strings");
const { domSymbolTree ***REMOVED*** = require("../helpers/internal-constants");
const { HTML_NS, SVG_NS ***REMOVED*** = require("../helpers/namespaces");
const { closest ***REMOVED*** = require("../helpers/traversal");
const { formOwner ***REMOVED*** = require("../helpers/form-controls");

class HTMLOptionElementImpl extends HTMLElementImpl {
  constructor(globalObject, args, privateData) {
    super(globalObject, args, privateData);

    // whenever selectedness is set to true, make sure all
    // other options set selectedness to false
    this._selectedness = false;
    this._dirtyness = false;
  ***REMOVED***

  _removeOtherSelectedness() {
    // Remove the selectedness flag from all other options in this select
    const select = this._selectNode;

    if (select && !select.hasAttributeNS(null, "multiple")) {
      for (const option of select.options) {
        if (option !== this) {
          option._selectedness = false;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  _askForAReset() {
    const select = this._selectNode;
    if (select) {
      select._askedForAReset();
    ***REMOVED***
  ***REMOVED***

  _attrModified(name, value, oldValue) {
    if (!this._dirtyness && name === "selected") {
      this._selectedness = this.hasAttributeNS(null, "selected");
      if (this._selectedness) {
        this._removeOtherSelectedness();
      ***REMOVED***
      this._askForAReset();
    ***REMOVED***
    super._attrModified(name, value, oldValue);
  ***REMOVED***

  get _selectNode() {
    let select = domSymbolTree.parent(this);
    if (!select) {
      return null;
    ***REMOVED***

    if (select.nodeName.toUpperCase() !== "SELECT") {
      select = domSymbolTree.parent(select);
      if (!select || select.nodeName.toUpperCase() !== "SELECT") {
        return null;
      ***REMOVED***
    ***REMOVED***
    return select;
  ***REMOVED***

  get form() {
    return formOwner(this);
  ***REMOVED***

  get text() {
    return stripAndCollapseASCIIWhitespace(childTextContentExcludingDescendantsOfScript(this));
  ***REMOVED***
  set text(value) {
    this.textContent = value;
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/form-elements.html#concept-option-value
  _getValue() {
    if (this.hasAttributeNS(null, "value")) {
      return this.getAttributeNS(null, "value");
    ***REMOVED***

    return this.text;
  ***REMOVED***

  get value() {
    return this._getValue();
  ***REMOVED***
  set value(value) {
    this.setAttributeNS(null, "value", value);
  ***REMOVED***

  get index() {
    const select = closest(this, "select");
    if (select === null) {
      return 0;
    ***REMOVED***

    return select.options.indexOf(this);
  ***REMOVED***

  get selected() {
    return this._selectedness;
  ***REMOVED***
  set selected(s) {
    this._dirtyness = true;
    this._selectedness = Boolean(s);
    if (this._selectedness) {
      this._removeOtherSelectedness();
    ***REMOVED***
    this._askForAReset();
    this._modified();
  ***REMOVED***

  get label() {
    if (this.hasAttributeNS(null, "label")) {
      return this.getAttributeNS(null, "label");
    ***REMOVED***

    return this.text;
  ***REMOVED***
  set label(value) {
    this.setAttributeNS(null, "label", value);
  ***REMOVED***
***REMOVED***

function childTextContentExcludingDescendantsOfScript(root) {
  let text = "";
  for (const child of domSymbolTree.childrenIterator(root)) {
    if (child._localName === "script" && (child._namespaceURI === HTML_NS || child._namespaceURI === SVG_NS)) {
      continue;
    ***REMOVED***

    if (child.nodeType === NODE_TYPE.TEXT_NODE || child.nodeType === NODE_TYPE.CDATA_SECTION_NODE) {
      text += child.nodeValue;
    ***REMOVED*** else {
      text += childTextContentExcludingDescendantsOfScript(child);
    ***REMOVED***
  ***REMOVED***
  return text;
***REMOVED***

module.exports = {
  implementation: HTMLOptionElementImpl
***REMOVED***;
