// jsdom/lib/jsdom/living/nodes/HTMLFormElement-impl.js
"use strict";

const DOMException = require("domexception/webidl2js-wrapper");
const { serializeURL ***REMOVED*** = require("whatwg-url");
const HTMLElementImpl = require("./HTMLElement-impl").implementation;
const { domSymbolTree ***REMOVED*** = require("../helpers/internal-constants");
const { fireAnEvent ***REMOVED*** = require("../helpers/events");
const { formOwner, isListed, isSubmittable, isSubmitButton ***REMOVED*** = require("../helpers/form-controls");
const HTMLCollection = require("../generated/HTMLCollection");
const notImplemented = require("../../browser/not-implemented");
const { parseURLToResultingURLRecord ***REMOVED*** = require("../helpers/document-base-url");

const encTypes = new Set([
  "application/x-www-form-urlencoded",
  "multipart/form-data",
  "text/plain"
]);

const methods = new Set([
  "get",
  "post",
  "dialog"
]);

const constraintValidationPositiveResult = Symbol("positive");
const constraintValidationNegativeResult = Symbol("negative");

class HTMLFormElementImpl extends HTMLElementImpl {
  _descendantAdded(parent, child) {
    const form = this;
    for (const el of domSymbolTree.treeIterator(child)) {
      if (typeof el._changedFormOwner === "function") {
        el._changedFormOwner(form);
      ***REMOVED***
    ***REMOVED***

    super._descendantAdded(parent, child);
  ***REMOVED***

  _descendantRemoved(parent, child) {
    for (const el of domSymbolTree.treeIterator(child)) {
      if (typeof el._changedFormOwner === "function") {
        el._changedFormOwner(null);
      ***REMOVED***
    ***REMOVED***

    super._descendantRemoved(parent, child);
  ***REMOVED***

  _getElementNodes() {
    return domSymbolTree.treeToArray(this.getRootNode({***REMOVED***), {
      filter: node => {
        if (!isListed(node) || (node._localName === "input" && node.type === "image")) {
          return false;
        ***REMOVED***

        return formOwner(node) === this;
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/forms.html#dom-form-elements
  get elements() {
    // TODO: Return a HTMLFormControlsCollection
    return HTMLCollection.createImpl(this._globalObject, [], {
      element: this.getRootNode({***REMOVED***),
      query: () => this._getElementNodes()
    ***REMOVED***);
  ***REMOVED***

  get length() {
    return this.elements.length;
  ***REMOVED***

  _doSubmit() {
    if (!this.isConnected) {
      return;
    ***REMOVED***

    this.submit();
  ***REMOVED***

  submit() {
    if (!fireAnEvent("submit", this, undefined, { bubbles: true, cancelable: true ***REMOVED***)) {
      return;
    ***REMOVED***

    notImplemented("HTMLFormElement.prototype.submit", this._ownerDocument._defaultView);
  ***REMOVED***

  requestSubmit(submitter = undefined) {
    if (submitter !== undefined) {
      if (!isSubmitButton(submitter)) {
        throw new TypeError("The specified element is not a submit button");
      ***REMOVED***
      if (submitter.form !== this) {
        throw DOMException.create(this._globalObject, [
          "The specified element is not owned by this form element",
          "NotFoundError"
        ]);
      ***REMOVED***
    ***REMOVED***

    if (!fireAnEvent("submit", this, undefined, { bubbles: true, cancelable: true ***REMOVED***)) {
      return;
    ***REMOVED***

    notImplemented("HTMLFormElement.prototype.requestSubmit", this._ownerDocument._defaultView);
  ***REMOVED***

  _doReset() {
    if (!this.isConnected) {
      return;
    ***REMOVED***

    this.reset();
  ***REMOVED***

  reset() {
    if (!fireAnEvent("reset", this, undefined, { bubbles: true, cancelable: true ***REMOVED***)) {
      return;
    ***REMOVED***

    for (const el of this.elements) {
      if (typeof el._formReset === "function") {
        el._formReset();
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  get method() {
    let method = this.getAttributeNS(null, "method");
    if (method) {
      method = method.toLowerCase();
    ***REMOVED***

    if (methods.has(method)) {
      return method;
    ***REMOVED***
    return "get";
  ***REMOVED***

  set method(V) {
    this.setAttributeNS(null, "method", V);
  ***REMOVED***

  get enctype() {
    let type = this.getAttributeNS(null, "enctype");
    if (type) {
      type = type.toLowerCase();
    ***REMOVED***

    if (encTypes.has(type)) {
      return type;
    ***REMOVED***
    return "application/x-www-form-urlencoded";
  ***REMOVED***

  set enctype(V) {
    this.setAttributeNS(null, "enctype", V);
  ***REMOVED***

  get action() {
    const attributeValue = this.getAttributeNS(null, "action");
    if (attributeValue === null || attributeValue === "") {
      return this._ownerDocument.URL;
    ***REMOVED***
    const urlRecord = parseURLToResultingURLRecord(attributeValue, this._ownerDocument);
    if (urlRecord === null) {
      return attributeValue;
    ***REMOVED***
    return serializeURL(urlRecord);
  ***REMOVED***

  set action(V) {
    this.setAttributeNS(null, "action", V);
  ***REMOVED***

  // If the checkValidity() method is invoked, the user agent must statically validate the
  // constraints of the form element, and return true if the constraint validation returned
  // a positive result, and false if it returned a negative result.
  checkValidity() {
    return this._staticallyValidateConstraints().result === constraintValidationPositiveResult;
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#interactively-validate-the-constraints
  reportValidity() {
    return this.checkValidity();
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#statically-validate-the-constraints
  _staticallyValidateConstraints() {
    const controls = [];
    for (const el of this.elements) {
      if (el.form === this && isSubmittable(el)) {
        controls.push(el);
      ***REMOVED***
    ***REMOVED***

    const invalidControls = [];

    for (const control of controls) {
      if (control._isCandidateForConstraintValidation() && !control._satisfiesConstraints()) {
        invalidControls.push(control);
      ***REMOVED***
    ***REMOVED***

    if (invalidControls.length === 0) {
      return { result: constraintValidationPositiveResult ***REMOVED***;
    ***REMOVED***

    const unhandledInvalidControls = [];
    for (const invalidControl of invalidControls) {
      const notCancelled = fireAnEvent("invalid", invalidControl, undefined, { cancelable: true ***REMOVED***);
      if (notCancelled) {
        unhandledInvalidControls.push(invalidControl);
      ***REMOVED***
    ***REMOVED***

    return { result: constraintValidationNegativeResult, unhandledInvalidControls ***REMOVED***;
  ***REMOVED***
***REMOVED***

module.exports = {
  implementation: HTMLFormElementImpl
***REMOVED***;
