// jsdom/lib/jsdom/living/nodes/HTMLAnchorElement-impl.js
"use strict";
const { mixin ***REMOVED*** = require("../../utils");
const DOMTokenList = require("../generated/DOMTokenList");
const HTMLElementImpl = require("./HTMLElement-impl").implementation;
const HTMLHyperlinkElementUtilsImpl = require("./HTMLHyperlinkElementUtils-impl").implementation;

class HTMLAnchorElementImpl extends HTMLElementImpl {
  constructor(globalObject, args, privateData) {
    super(globalObject, args, privateData);

    this._htmlHyperlinkElementUtilsSetup();

    this._hasActivationBehavior = true;
  ***REMOVED***

  _activationBehavior() {
    this._followAHyperlink();
  ***REMOVED***

  get relList() {
    if (this._relList === undefined) {
      this._relList = DOMTokenList.createImpl(this._globalObject, [], {
        element: this,
        attributeLocalName: "rel"
      ***REMOVED***);
    ***REMOVED***
    return this._relList;
  ***REMOVED***

  get text() {
    return this.textContent;
  ***REMOVED***
  set text(v) {
    this.textContent = v;
  ***REMOVED***

  _attrModified(name, value, oldValue) {
    super._attrModified(name, value, oldValue);

    if (name === "rel" && this._relList !== undefined) {
      this._relList.attrModified();
    ***REMOVED***
  ***REMOVED***
***REMOVED***

mixin(HTMLAnchorElementImpl.prototype, HTMLHyperlinkElementUtilsImpl.prototype);

module.exports = {
  implementation: HTMLAnchorElementImpl
***REMOVED***;
