// jsdom/lib/jsdom/living/nodes/HTMLTableSectionElement-impl.js
"use strict";

const HTMLElementImpl = require("./HTMLElement-impl").implementation;
const { childrenByLocalName ***REMOVED*** = require("../helpers/traversal");
const HTMLCollection = require("../generated/HTMLCollection");
const DOMException = require("domexception/webidl2js-wrapper");

class HTMLTableSectionElementImpl extends HTMLElementImpl {
  get rows() {
    if (!this._rows) {
      this._rows = HTMLCollection.createImpl(this._globalObject, [], {
        element: this,
        query: () => childrenByLocalName(this, "tr")
      ***REMOVED***);
    ***REMOVED***
    return this._rows;
  ***REMOVED***

  insertRow(index) {
    if (index < -1 || index > this.rows.length) {
      throw DOMException.create(this._globalObject, [
        "Cannot insert a row at an index that is less than -1 or greater than the number of existing rows",
        "IndexSizeError"
      ]);
    ***REMOVED***

    const tr = this._ownerDocument.createElement("tr");

    if (index === -1 || index === this.rows.length) {
      this._append(tr);
    ***REMOVED*** else {
      const beforeTR = this.rows.item(index);
      this._insert(tr, beforeTR);
    ***REMOVED***

    return tr;
  ***REMOVED***

  deleteRow(index) {
    if (index < -1 || index >= this.rows.length) {
      throw DOMException.create(this._globalObject, [
        `Cannot delete a row at index ${index***REMOVED***, where no row exists`,
        "IndexSizeError"
      ]);
    ***REMOVED***

    if (index === -1) {
      if (this.rows.length > 0) {
        const tr = this.rows.item(this.rows.length - 1);
        this._remove(tr);
      ***REMOVED***
    ***REMOVED*** else {
      const tr = this.rows.item(index);
      this._remove(tr);
    ***REMOVED***
  ***REMOVED***
***REMOVED***

module.exports = {
  implementation: HTMLTableSectionElementImpl
***REMOVED***;
