// jsdom/lib/jsdom/living/nodes/HTMLTableCellElement-impl.js
"use strict";

const HTMLElementImpl = require("./HTMLElement-impl").implementation;

const { asciiLowercase, parseNonNegativeInteger ***REMOVED*** = require("../helpers/strings");
const { closest ***REMOVED*** = require("../helpers/traversal");

function reflectedAttributeClampedToRange(attrValue, min, max, defaultValue = 0) {
  if (attrValue === null) {
    return defaultValue;
  ***REMOVED***
  const parsed = parseNonNegativeInteger(attrValue);
  if (parsed === null) {
    return defaultValue;
  ***REMOVED***
  if (parsed < min) {
    return min;
  ***REMOVED***
  if (parsed > max) {
    return max;
  ***REMOVED***
  return parsed;
***REMOVED***

class HTMLTableCellElementImpl extends HTMLElementImpl {
  get colSpan() {
    return reflectedAttributeClampedToRange(this.getAttributeNS(null, "colspan"), 1, 1000, 1);
  ***REMOVED***

  set colSpan(V) {
    this.setAttributeNS(null, "colspan", String(V));
  ***REMOVED***

  get rowSpan() {
    return reflectedAttributeClampedToRange(this.getAttributeNS(null, "rowspan"), 0, 65534, 1);
  ***REMOVED***

  set rowSpan(V) {
    this.setAttributeNS(null, "rowspan", String(V));
  ***REMOVED***

  get cellIndex() {
    const tr = closest(this, "tr");
    if (tr === null) {
      return -1;
    ***REMOVED***

    return tr.cells.indexOf(this);
  ***REMOVED***

  get scope() {
    let value = this.getAttributeNS(null, "scope");
    if (value === null) {
      return "";
    ***REMOVED***

    // Enumerated attribute is matched ASCII-case-insensitively.
    value = asciiLowercase(value);
    if (value === "row" || value === "col" || value === "rowgroup" || value === "colgroup") {
      return value;
    ***REMOVED***

    return "";
  ***REMOVED***

  set scope(V) {
    this.setAttributeNS(null, "scope", V);
  ***REMOVED***
***REMOVED***

module.exports = {
  implementation: HTMLTableCellElementImpl
***REMOVED***;
