// jsdom/lib/jsdom/living/nodes/HTMLImageElement-impl.js
"use strict";
const conversions = require("webidl-conversions");
const { serializeURL ***REMOVED*** = require("whatwg-url");
const HTMLElementImpl = require("./HTMLElement-impl").implementation;
const { Canvas ***REMOVED*** = require("../../utils");
const { parseURLToResultingURLRecord ***REMOVED*** = require("../helpers/document-base-url");

class HTMLImageElementImpl extends HTMLElementImpl {
  constructor(...args) {
    super(...args);
    this._currentRequestState = "unavailable";
  ***REMOVED***

  _attrModified(name, value, oldVal) {
    // TODO: handle crossorigin
    if (name === "src" || ((name === "srcset" || name === "width" || name === "sizes") && value !== oldVal)) {
      this._updateTheImageData();
    ***REMOVED***

    super._attrModified(name, value, oldVal);
  ***REMOVED***

  get _accept() {
    return "image/png,image/*;q=0.8,*/*;q=0.5";
  ***REMOVED***

  get height() {
    // Just like on browsers, if no width / height is defined, we fall back on the
    // dimensions of the internal image data.
    return this.hasAttributeNS(null, "height") ?
           conversions["unsigned long"](this.getAttributeNS(null, "height")) :
           this.naturalHeight;
  ***REMOVED***

  set height(V) {
    this.setAttributeNS(null, "height", String(V));
  ***REMOVED***

  get width() {
    return this.hasAttributeNS(null, "width") ?
           conversions["unsigned long"](this.getAttributeNS(null, "width")) :
           this.naturalWidth;
  ***REMOVED***

  set width(V) {
    this.setAttributeNS(null, "width", String(V));
  ***REMOVED***

  get naturalHeight() {
    return this._image ? this._image.naturalHeight : 0;
  ***REMOVED***

  get naturalWidth() {
    return this._image ? this._image.naturalWidth : 0;
  ***REMOVED***

  get complete() {
    const srcAttributeValue = this.getAttributeNS(null, "src");
    return srcAttributeValue === null ||
      srcAttributeValue === "" ||
      this._currentRequestState === "broken" ||
      this._currentRequestState === "completely available";
  ***REMOVED***

  get currentSrc() {
    return this._currentSrc || "";
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/images.html#updating-the-image-data
  _updateTheImageData() {
    const document = this._ownerDocument;

    if (!document._defaultView) {
      return;
    ***REMOVED***

    if (!Canvas) {
      return;
    ***REMOVED***

    if (!this._image) {
      this._image = new Canvas.Image();
    ***REMOVED***
    this._currentSrc = null;
    this._currentRequestState = "unavailable";
    const srcAttributeValue = this.getAttributeNS(null, "src");
    let urlString = null;
    if (srcAttributeValue !== null && srcAttributeValue !== "") {
      const urlRecord = parseURLToResultingURLRecord(srcAttributeValue, this._ownerDocument);
      if (urlRecord === null) {
        return;
      ***REMOVED***
      urlString = serializeURL(urlRecord);
    ***REMOVED***
    if (urlString !== null) {
      const resourceLoader = document._resourceLoader;
      let request;

      const onLoadImage = data => {
        const { response ***REMOVED*** = request;

        if (response && response.statusCode !== undefined && response.statusCode !== 200) {
          throw new Error("Status code: " + response.statusCode);
        ***REMOVED***
        let error = null;
        this._image.onerror = function (err) {
          error = err;
        ***REMOVED***;
        this._image.src = data;
        if (error) {
          throw new Error(error);
        ***REMOVED***
        this._currentSrc = srcAttributeValue;
        this._currentRequestState = "completely available";
      ***REMOVED***;

      request = resourceLoader.fetch(urlString, {
        element: this,
        onLoad: onLoadImage,
        onError: () => {
          this._currentRequestState = "broken";
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED*** else {
      this._image.src = "";
    ***REMOVED***
  ***REMOVED***
***REMOVED***

module.exports = {
  implementation: HTMLImageElementImpl
***REMOVED***;
