// jsdom/lib/jsdom/living/nodes/HTMLMeterElement-impl.js
"use strict";

const HTMLElementImpl = require("./HTMLElement-impl").implementation;
const { parseFloatingPointNumber ***REMOVED*** = require("../helpers/strings");
const { getLabelsForLabelable ***REMOVED*** = require("../helpers/form-controls");

class HTMLMeterElementImpl extends HTMLElementImpl {
  constructor(globalObject, args, privateData) {
    super(globalObject, args, privateData);
    this._labels = null;
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/form-elements.html#concept-meter-minimum
  get _minimumValue() {
    const min = this.getAttributeNS(null, "min");
    if (min !== null) {
      const parsed = parseFloatingPointNumber(min);
      if (parsed !== null) {
        return parsed;
      ***REMOVED***
    ***REMOVED***
    return 0;
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/form-elements.html#concept-meter-maximum
  get _maximumValue() {
    let candidate = 1.0;

    const max = this.getAttributeNS(null, "max");
    if (max !== null) {
      const parsed = parseFloatingPointNumber(max);
      if (parsed !== null) {
        candidate = parsed;
      ***REMOVED***
    ***REMOVED***

    const minimumValue = this._minimumValue;
    return candidate >= minimumValue ? candidate : minimumValue;
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/form-elements.html#concept-meter-actual
  get _actualValue() {
    let candidate = 0;

    const value = this.getAttributeNS(null, "value");
    if (value !== null) {
      const parsed = parseFloatingPointNumber(value);
      if (parsed !== null) {
        candidate = parsed;
      ***REMOVED***
    ***REMOVED***

    const minimumValue = this._minimumValue;
    if (candidate < minimumValue) {
      return minimumValue;
    ***REMOVED***

    const maximumValue = this._maximumValue;
    return candidate > maximumValue ? maximumValue : candidate;
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/form-elements.html#concept-meter-low
  get _lowBoundary() {
    const minimumValue = this._minimumValue;
    let candidate = minimumValue;

    const low = this.getAttributeNS(null, "low");
    if (low !== null) {
      const parsed = parseFloatingPointNumber(low);
      if (parsed !== null) {
        candidate = parsed;
      ***REMOVED***
    ***REMOVED***

    if (candidate < minimumValue) {
      return minimumValue;
    ***REMOVED***

    const maximumValue = this._maximumValue;
    return candidate > maximumValue ? maximumValue : candidate;
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/form-elements.html#concept-meter-high
  get _highBoundary() {
    const maximumValue = this._maximumValue;
    let candidate = maximumValue;

    const high = this.getAttributeNS(null, "high");
    if (high !== null) {
      const parsed = parseFloatingPointNumber(high);
      if (parsed !== null) {
        candidate = parsed;
      ***REMOVED***
    ***REMOVED***

    const lowBoundary = this._lowBoundary;
    if (candidate < lowBoundary) {
      return lowBoundary;
    ***REMOVED***

    return candidate > maximumValue ? maximumValue : candidate;
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/form-elements.html#concept-meter-optimum
  get _optimumPoint() {
    const minimumValue = this._minimumValue;
    const maximumValue = this._maximumValue;
    let candidate = (minimumValue + maximumValue) / 2;

    const optimum = this.getAttributeNS(null, "optimum");
    if (optimum !== null) {
      const parsed = parseFloatingPointNumber(optimum);
      if (parsed !== null) {
        candidate = parsed;
      ***REMOVED***
    ***REMOVED***

    if (candidate < minimumValue) {
      return minimumValue;
    ***REMOVED***

    return candidate > maximumValue ? maximumValue : candidate;
  ***REMOVED***

  get labels() {
    return getLabelsForLabelable(this);
  ***REMOVED***

  get value() {
    return this._actualValue;
  ***REMOVED***

  set value(val) {
    this.setAttributeNS(null, "value", String(val));
  ***REMOVED***

  get min() {
    return this._minimumValue;
  ***REMOVED***

  set min(val) {
    this.setAttributeNS(null, "min", String(val));
  ***REMOVED***

  get max() {
    return this._maximumValue;
  ***REMOVED***

  set max(val) {
    this.setAttributeNS(null, "max", String(val));
  ***REMOVED***

  get low() {
    return this._lowBoundary;
  ***REMOVED***

  set low(val) {
    this.setAttributeNS(null, "low", String(val));
  ***REMOVED***

  get high() {
    return this._highBoundary;
  ***REMOVED***

  set high(val) {
    this.setAttributeNS(null, "high", String(val));
  ***REMOVED***

  get optimum() {
    return this._optimumPoint;
  ***REMOVED***

  set optimum(val) {
    this.setAttributeNS(null, "optimum", String(val));
  ***REMOVED***
***REMOVED***

module.exports = {
  implementation: HTMLMeterElementImpl
***REMOVED***;
