// jsdom/lib/jsdom/living/nodes/HTMLOutputElement-impl.js
"use strict";

const DOMTokenList = require("../generated/DOMTokenList");
const HTMLElementImpl = require("./HTMLElement-impl").implementation;
const DefaultConstraintValidationImpl =
  require("../constraint-validation/DefaultConstraintValidation-impl").implementation;
const { mixin ***REMOVED*** = require("../../utils");
const { getLabelsForLabelable, formOwner ***REMOVED*** = require("../helpers/form-controls");

class HTMLOutputElementImpl extends HTMLElementImpl {
  constructor(globalObject, args, privateData) {
    super(globalObject, args, privateData);
    this._labels = null;
    this._defaultValueOverride = null;

    this._customValidityErrorMessage = "";
  ***REMOVED***

  _attrModified(name, value, oldValue) {
    super._attrModified(name, value, oldValue);

    if (name === "for" && this._htmlFor !== undefined) {
      this._htmlFor.attrModified();
    ***REMOVED***
  ***REMOVED***

  _barredFromConstraintValidationSpecialization() {
    return true;
  ***REMOVED***

  _formReset() {
    this.textContent = this.defaultValue;
    this._defaultValueOverride = null;
  ***REMOVED***

  get htmlFor() {
    if (this._htmlFor === undefined) {
      this._htmlFor = DOMTokenList.createImpl(this._globalObject, [], {
        element: this,
        attributeLocalName: "for"
      ***REMOVED***);
    ***REMOVED***
    return this._htmlFor;
  ***REMOVED***

  get type() {
    return "output";
  ***REMOVED***

  get labels() {
    return getLabelsForLabelable(this);
  ***REMOVED***

  get form() {
    return formOwner(this);
  ***REMOVED***

  get value() {
    return this.textContent;
  ***REMOVED***

  set value(val) {
    this._defaultValueOverride = this.defaultValue;
    this.textContent = val;
  ***REMOVED***

  get defaultValue() {
    if (this._defaultValueOverride !== null) {
      return this._defaultValueOverride;
    ***REMOVED***
    return this.textContent;
  ***REMOVED***

  set defaultValue(val) {
    if (this._defaultValueOverride === null) {
      this.textContent = val;
      return;
    ***REMOVED***

    this._defaultValueOverride = val;
  ***REMOVED***
***REMOVED***

mixin(HTMLOutputElementImpl.prototype, DefaultConstraintValidationImpl.prototype);

module.exports = {
  implementation: HTMLOutputElementImpl
***REMOVED***;
