// jsdom/lib/jsdom/living/nodes/HTMLTemplateElement-impl.js
"use strict";

const HTMLElementImpl = require("./HTMLElement-impl").implementation;

const Document = require("../generated/Document");
const DocumentFragment = require("../generated/DocumentFragment");

const { cloningSteps, domSymbolTree ***REMOVED*** = require("../helpers/internal-constants");
const { clone ***REMOVED*** = require("../node");

class HTMLTemplateElementImpl extends HTMLElementImpl {
  constructor(globalObject, args, privateData) {
    super(globalObject, args, privateData);

    const doc = this._appropriateTemplateContentsOwnerDocument(this._ownerDocument);
    this._templateContents = DocumentFragment.createImpl(this._globalObject, [], {
      ownerDocument: doc,
      host: this
    ***REMOVED***);
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/scripting.html#appropriate-template-contents-owner-document
  _appropriateTemplateContentsOwnerDocument(doc) {
    if (!doc._isInertTemplateDocument) {
      if (doc._associatedInertTemplateDocument === undefined) {
        const newDoc = Document.createImpl(this._globalObject, [], {
          options: {
            parsingMode: doc._parsingMode,
            encoding: doc._encoding
          ***REMOVED***
        ***REMOVED***);
        newDoc._isInertTemplateDocument = true;

        doc._associatedInertTemplateDocument = newDoc;
      ***REMOVED***

      doc = doc._associatedInertTemplateDocument;
    ***REMOVED***

    return doc;
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/scripting.html#template-adopting-steps
  _adoptingSteps() {
    const doc = this._appropriateTemplateContentsOwnerDocument(this._ownerDocument);
    doc._adoptNode(this._templateContents);
  ***REMOVED***

  get content() {
    return this._templateContents;
  ***REMOVED***

  [cloningSteps](copy, node, document, cloneChildren) {
    if (!cloneChildren) {
      return;
    ***REMOVED***

    for (const child of domSymbolTree.childrenIterator(node._templateContents)) {
      const childCopy = clone(child, copy._templateContents._ownerDocument, true);
      copy._templateContents.appendChild(childCopy);
    ***REMOVED***
  ***REMOVED***
***REMOVED***

module.exports = {
  implementation: HTMLTemplateElementImpl
***REMOVED***;
