// jsdom/lib/jsdom/living/nodes/HTMLFieldSetElement-impl.js
"use strict";
const HTMLCollection = require("../generated/HTMLCollection");
const HTMLElementImpl = require("./HTMLElement-impl").implementation;
const DefaultConstraintValidationImpl =
  require("../constraint-validation/DefaultConstraintValidation-impl").implementation;
const { formOwner ***REMOVED*** = require("../helpers/form-controls");
const { mixin ***REMOVED*** = require("../../utils");
const { descendantsByLocalNames ***REMOVED*** = require("../helpers/traversal");

const listedElements = new Set(["button", "fieldset", "input", "object", "output", "select", "textarea"]);

class HTMLFieldSetElementImpl extends HTMLElementImpl {
  constructor(globalObject, args, privateData) {
    super(globalObject, args, privateData);

    this._customValidityErrorMessage = "";
  ***REMOVED***

  get elements() {
    return HTMLCollection.createImpl(this._globalObject, [], {
      element: this,
      query: () => descendantsByLocalNames(this, listedElements)
    ***REMOVED***);
  ***REMOVED***

  get form() {
    return formOwner(this);
  ***REMOVED***

  get type() {
    return "fieldset";
  ***REMOVED***

  _barredFromConstraintValidationSpecialization() {
    return true;
  ***REMOVED***
***REMOVED***

mixin(HTMLFieldSetElementImpl.prototype, DefaultConstraintValidationImpl.prototype);

module.exports = {
  implementation: HTMLFieldSetElementImpl
***REMOVED***;
