// jsdom/lib/jsdom/living/nodes/HTMLHyperlinkElementUtils-impl.js
"use strict";
const whatwgURL = require("whatwg-url");
const { parseURLToResultingURLRecord ***REMOVED*** = require("../helpers/document-base-url");
const { asciiCaseInsensitiveMatch ***REMOVED*** = require("../helpers/strings");
const { navigate ***REMOVED*** = require("../window/navigation");

exports.implementation = class HTMLHyperlinkElementUtilsImpl {
  _htmlHyperlinkElementUtilsSetup() {
    this.url = null;
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/links.html#cannot-navigate
  _cannotNavigate() {
    // TODO: Correctly check if the document is fully active
    return this._localName !== "a" && !this.isConnected;
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/semantics.html#get-an-element's-target
  _getAnElementsTarget() {
    if (this.hasAttributeNS(null, "target")) {
      return this.getAttributeNS(null, "target");
    ***REMOVED***

    const baseEl = this._ownerDocument.querySelector("base[target]");

    if (baseEl) {
      return baseEl.getAttributeNS(null, "target");
    ***REMOVED***

    return "";
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/browsers.html#the-rules-for-choosing-a-browsing-context-given-a-browsing-context-name
  _chooseABrowsingContext(name, current) {
    let chosen = null;

    if (name === "" || asciiCaseInsensitiveMatch(name, "_self")) {
      chosen = current;
    ***REMOVED*** else if (asciiCaseInsensitiveMatch(name, "_parent")) {
      chosen = current.parent;
    ***REMOVED*** else if (asciiCaseInsensitiveMatch(name, "_top")) {
      chosen = current.top;
    ***REMOVED*** else if (!asciiCaseInsensitiveMatch(name, "_blank")) {
      // https://github.com/whatwg/html/issues/1440
    ***REMOVED***

    // TODO: Create new browsing context, handle noopener

    return chosen;
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/links.html#following-hyperlinks-2
  _followAHyperlink() {
    if (this._cannotNavigate()) {
      return;
    ***REMOVED***

    const source = this._ownerDocument._defaultView;
    let targetAttributeValue = "";

    if (this._localName === "a" || this._localName === "area") {
      targetAttributeValue = this._getAnElementsTarget();
    ***REMOVED***

    const noopener = this.relList.contains("noreferrer") || this.relList.contains("noopener");

    const target = this._chooseABrowsingContext(targetAttributeValue, source, noopener);

    if (target === null) {
      return;
    ***REMOVED***

    const url = parseURLToResultingURLRecord(this.href, this._ownerDocument);

    if (url === null) {
      return;
    ***REMOVED***

    // TODO: Handle hyperlink suffix and referrerpolicy
    setTimeout(() => {
      navigate(target, url, {***REMOVED***);
    ***REMOVED***, 0);
  ***REMOVED***

  toString() {
    return this.href;
  ***REMOVED***

  get href() {
    reinitializeURL(this);
    const { url ***REMOVED*** = this;

    if (url === null) {
      const href = this.getAttributeNS(null, "href");
      return href === null ? "" : href;
    ***REMOVED***

    return whatwgURL.serializeURL(url);
  ***REMOVED***

  set href(v) {
    this.setAttributeNS(null, "href", v);
  ***REMOVED***

  get origin() {
    reinitializeURL(this);

    if (this.url === null) {
      return "";
    ***REMOVED***

    return whatwgURL.serializeURLOrigin(this.url);
  ***REMOVED***

  get protocol() {
    reinitializeURL(this);

    if (this.url === null) {
      return ":";
    ***REMOVED***

    return this.url.scheme + ":";
  ***REMOVED***

  set protocol(v) {
    reinitializeURL(this);

    if (this.url === null) {
      return;
    ***REMOVED***

    whatwgURL.basicURLParse(v + ":", { url: this.url, stateOverride: "scheme start" ***REMOVED***);
    updateHref(this);
  ***REMOVED***

  get username() {
    reinitializeURL(this);

    if (this.url === null) {
      return "";
    ***REMOVED***

    return this.url.username;
  ***REMOVED***

  set username(v) {
    reinitializeURL(this);
    const { url ***REMOVED*** = this;

    if (url === null || url.host === null || url.host === "" || url.cannotBeABaseURL || url.scheme === "file") {
      return;
    ***REMOVED***

    whatwgURL.setTheUsername(url, v);
    updateHref(this);
  ***REMOVED***

  get password() {
    reinitializeURL(this);
    const { url ***REMOVED*** = this;

    if (url === null) {
      return "";
    ***REMOVED***

    return url.password;
  ***REMOVED***

  set password(v) {
    reinitializeURL(this);
    const { url ***REMOVED*** = this;

    if (url === null || url.host === null || url.host === "" || url.cannotBeABaseURL || url.scheme === "file") {
      return;
    ***REMOVED***

    whatwgURL.setThePassword(url, v);
    updateHref(this);
  ***REMOVED***

  get host() {
    reinitializeURL(this);
    const { url ***REMOVED*** = this;

    if (url === null || url.host === null) {
      return "";
    ***REMOVED***

    if (url.port === null) {
      return whatwgURL.serializeHost(url.host);
    ***REMOVED***

    return whatwgURL.serializeHost(url.host) + ":" + whatwgURL.serializeInteger(url.port);
  ***REMOVED***

  set host(v) {
    reinitializeURL(this);
    const { url ***REMOVED*** = this;

    if (url === null || url.cannotBeABaseURL) {
      return;
    ***REMOVED***

    whatwgURL.basicURLParse(v, { url, stateOverride: "host" ***REMOVED***);
    updateHref(this);
  ***REMOVED***

  get hostname() {
    reinitializeURL(this);
    const { url ***REMOVED*** = this;

    if (url === null || url.host === null) {
      return "";
    ***REMOVED***

    return whatwgURL.serializeHost(url.host);
  ***REMOVED***

  set hostname(v) {
    reinitializeURL(this);
    const { url ***REMOVED*** = this;

    if (url === null || url.cannotBeABaseURL) {
      return;
    ***REMOVED***

    whatwgURL.basicURLParse(v, { url, stateOverride: "hostname" ***REMOVED***);
    updateHref(this);
  ***REMOVED***

  get port() {
    reinitializeURL(this);
    const { url ***REMOVED*** = this;

    if (url === null || url.port === null) {
      return "";
    ***REMOVED***

    return whatwgURL.serializeInteger(url.port);
  ***REMOVED***

  set port(v) {
    reinitializeURL(this);
    const { url ***REMOVED*** = this;

    if (url === null || url.host === null || url.host === "" || url.cannotBeABaseURL || url.scheme === "file") {
      return;
    ***REMOVED***

    if (v === "") {
      url.port = null;
    ***REMOVED*** else {
      whatwgURL.basicURLParse(v, { url, stateOverride: "port" ***REMOVED***);
    ***REMOVED***
    updateHref(this);
  ***REMOVED***

  get pathname() {
    reinitializeURL(this);
    const { url ***REMOVED*** = this;

    if (url === null) {
      return "";
    ***REMOVED***

    if (url.cannotBeABaseURL) {
      return url.path[0];
    ***REMOVED***

    return "/" + url.path.join("/");
  ***REMOVED***

  set pathname(v) {
    reinitializeURL(this);
    const { url ***REMOVED*** = this;

    if (url === null || url.cannotBeABaseURL) {
      return;
    ***REMOVED***

    url.path = [];
    whatwgURL.basicURLParse(v, { url, stateOverride: "path start" ***REMOVED***);
    updateHref(this);
  ***REMOVED***

  get search() {
    reinitializeURL(this);
    const { url ***REMOVED*** = this;

    if (url === null || url.query === null || url.query === "") {
      return "";
    ***REMOVED***

    return "?" + url.query;
  ***REMOVED***

  set search(v) {
    reinitializeURL(this);
    const { url ***REMOVED*** = this;

    if (url === null) {
      return;
    ***REMOVED***

    if (v === "") {
      url.query = null;
    ***REMOVED*** else {
      const input = v[0] === "?" ? v.substring(1) : v;
      url.query = "";
      whatwgURL.basicURLParse(input, {
        url,
        stateOverride: "query",
        encodingOverride: this._ownerDocument.charset
      ***REMOVED***);
    ***REMOVED***
    updateHref(this);
  ***REMOVED***

  get hash() {
    reinitializeURL(this);
    const { url ***REMOVED*** = this;

    if (url === null || url.fragment === null || url.fragment === "") {
      return "";
    ***REMOVED***

    return "#" + url.fragment;
  ***REMOVED***

  set hash(v) {
    reinitializeURL(this);
    const { url ***REMOVED*** = this;

    if (url === null) {
      return;
    ***REMOVED***

    if (v === "") {
      url.fragment = null;
    ***REMOVED*** else {
      const input = v[0] === "#" ? v.substring(1) : v;
      url.fragment = "";
      whatwgURL.basicURLParse(input, { url, stateOverride: "fragment" ***REMOVED***);
    ***REMOVED***
    updateHref(this);
  ***REMOVED***
***REMOVED***;

function reinitializeURL(hheu) {
  if (hheu.url !== null && hheu.url.scheme === "blob" && hheu.url.cannotBeABaseURL) {
    return;
  ***REMOVED***

  setTheURL(hheu);
***REMOVED***

function setTheURL(hheu) {
  const href = hheu.getAttributeNS(null, "href");
  if (href === null) {
    hheu.url = null;
    return;
  ***REMOVED***

  const parsed = parseURLToResultingURLRecord(href, hheu._ownerDocument);

  hheu.url = parsed === null ? null : parsed;
***REMOVED***

function updateHref(hheu) {
  hheu.setAttributeNS(null, "href", whatwgURL.serializeURL(hheu.url));
***REMOVED***
