// jsdom/lib/jsdom/living/nodes/HTMLElement-impl.js
"use strict";
const { mixin ***REMOVED*** = require("../../utils");
const ElementImpl = require("./Element-impl").implementation;
const MouseEvent = require("../generated/MouseEvent");
const ElementCSSInlineStyleImpl = require("./ElementCSSInlineStyle-impl").implementation;
const GlobalEventHandlersImpl = require("./GlobalEventHandlers-impl").implementation;
const HTMLOrSVGElementImpl = require("./HTMLOrSVGElement-impl").implementation;
const { firstChildWithLocalName ***REMOVED*** = require("../helpers/traversal");
const { isDisabled ***REMOVED*** = require("../helpers/form-controls");
const { fireAnEvent ***REMOVED*** = require("../helpers/events");
const { asciiLowercase ***REMOVED*** = require("../helpers/strings");

class HTMLElementImpl extends ElementImpl {
  constructor(globalObject, args, privateData) {
    super(globalObject, args, privateData);
    this._initHTMLOrSVGElement();
    this._initElementCSSInlineStyle();
    this._initGlobalEvents();

    this._clickInProgress = false;

    // <summary> uses HTMLElement and has activation behavior
    this._hasActivationBehavior = this._localName === "summary";
  ***REMOVED***

  _activationBehavior() {
    const parent = this.parentNode;
    if (parent && parent._localName === "details" &&
        this === firstChildWithLocalName(parent, "summary")) {
      if (parent.hasAttributeNS(null, "open")) {
        parent.removeAttributeNS(null, "open");
      ***REMOVED*** else {
        parent.setAttributeNS(null, "open", "");
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/dom.html#the-translate-attribute
  get translate() {
    const translateAttr = this.getAttributeNS(null, "translate");
    const translateAttrString = asciiLowercase(translateAttr || "");

    if (translateAttrString === "yes" || (translateAttr && translateAttrString === "")) {
      return true;
    ***REMOVED*** else if (translateAttrString === "no") {
      return false;
    ***REMOVED***

    if (this === this.ownerDocument.documentElement) {
      return true;
    ***REMOVED***

    return this.parentElement && this.parentElement.translate;
  ***REMOVED***
  set translate(value) {
    if (value === true) {
      this.setAttributeNS(null, "translate", "yes");
    ***REMOVED*** else {
      this.setAttributeNS(null, "translate", "no");
    ***REMOVED***
  ***REMOVED***

  click() {
    // https://html.spec.whatwg.org/multipage/interaction.html#dom-click
    // https://html.spec.whatwg.org/multipage/webappapis.html#fire-a-synthetic-mouse-event

    if (isDisabled(this)) {
      return;
    ***REMOVED***

    if (this._clickInProgress) {
      return;
    ***REMOVED***

    this._clickInProgress = true;

    // https://github.com/whatwg/html/issues/4451
    // https://github.com/whatwg/html/issues/4452
    fireAnEvent("click", this, MouseEvent, {
      bubbles: true,
      cancelable: true,
      composed: true,
      isTrusted: false,
      view: this.ownerDocument.defaultView
    ***REMOVED***);

    this._clickInProgress = false;
  ***REMOVED***

  get draggable() {
    const attributeValue = asciiLowercase(this.getAttributeNS(null, "draggable") || "");

    if (attributeValue === "true") {
      return true;
    ***REMOVED*** else if (attributeValue === "false") {
      return false;
    ***REMOVED***

    return this._localName === "img" || (this._localName === "a" && this.hasAttributeNS(null, "href"));
  ***REMOVED***
  set draggable(value) {
    this.setAttributeNS(null, "draggable", String(value));
  ***REMOVED***

  get dir() {
    let dirValue = this.getAttributeNS(null, "dir");
    if (dirValue !== null) {
      dirValue = dirValue.toLowerCase();

      if (["ltr", "rtl", "auto"].includes(dirValue)) {
        return dirValue;
      ***REMOVED***
    ***REMOVED***
    return "";
  ***REMOVED***
  set dir(value) {
    this.setAttributeNS(null, "dir", value);
  ***REMOVED***

  // Keep in sync with SVGElement. https://github.com/jsdom/jsdom/issues/2599
  _attrModified(name, value, oldValue) {
    if (name === "style" && value !== oldValue && !this._settingCssText) {
      this._settingCssText = true;
      this._style.cssText = value;
      this._settingCssText = false;
    ***REMOVED*** else if (name.startsWith("on")) {
      this._globalEventChanged(name.substring(2));
    ***REMOVED***

    super._attrModified(name, value, oldValue);
  ***REMOVED***

  get offsetParent() {
    return null;
  ***REMOVED***

  get offsetTop() {
    return 0;
  ***REMOVED***

  get offsetLeft() {
    return 0;
  ***REMOVED***

  get offsetWidth() {
    return 0;
  ***REMOVED***

  get offsetHeight() {
    return 0;
  ***REMOVED***
***REMOVED***

mixin(HTMLElementImpl.prototype, ElementCSSInlineStyleImpl.prototype);
mixin(HTMLElementImpl.prototype, GlobalEventHandlersImpl.prototype);
mixin(HTMLElementImpl.prototype, HTMLOrSVGElementImpl.prototype);

module.exports = {
  implementation: HTMLElementImpl
***REMOVED***;
