// jsdom/lib/jsdom/living/nodes/HTMLTextAreaElement-impl.js
"use strict";

const HTMLElementImpl = require("./HTMLElement-impl").implementation;

const DefaultConstraintValidationImpl =
  require("../constraint-validation/DefaultConstraintValidation-impl").implementation;
const ValidityState = require("../generated/ValidityState");
const { mixin ***REMOVED*** = require("../../utils");

const DOMException = require("domexception/webidl2js-wrapper");
const { cloningSteps ***REMOVED*** = require("../helpers/internal-constants");
const { isDisabled, getLabelsForLabelable, formOwner ***REMOVED*** = require("../helpers/form-controls");
const { childTextContent ***REMOVED*** = require("../helpers/text");
const { fireAnEvent ***REMOVED*** = require("../helpers/events");

class HTMLTextAreaElementImpl extends HTMLElementImpl {
  constructor(globalObject, args, privateData) {
    super(globalObject, args, privateData);

    this._selectionStart = this._selectionEnd = 0;
    this._selectionDirection = "none";
    this._rawValue = "";
    this._dirtyValue = false;

    this._customValidityErrorMessage = "";

    this._labels = null;
  ***REMOVED***

  _formReset() {
    this._rawValue = childTextContent(this);
    this._dirtyValue = false;
  ***REMOVED***

  _getAPIValue() {
    return this._rawValue.replace(/\r\n/g, "\n").replace(/\r/g, "\n");
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/form-elements.html#textarea-wrapping-transformation
  _getValue() {
    const apiValue = this._getAPIValue();
    const wrap = this.getAttributeNS(null, "wrap");
    return wrap === "hard" ?
      textareaWrappingTransformation(apiValue, this.cols) :
      apiValue;
  ***REMOVED***

  _childTextContentChangeSteps() {
    super._childTextContentChangeSteps();

    if (this._dirtyValue === false) {
      this._rawValue = childTextContent(this);
    ***REMOVED***
  ***REMOVED***

  get labels() {
    return getLabelsForLabelable(this);
  ***REMOVED***

  get form() {
    return formOwner(this);
  ***REMOVED***

  get defaultValue() {
    return childTextContent(this);
  ***REMOVED***

  set defaultValue(val) {
    this.textContent = val;
  ***REMOVED***

  get value() {
    return this._getAPIValue();
  ***REMOVED***

  set value(val) {
    // https://html.spec.whatwg.org/multipage/form-elements.html#dom-textarea-value
    const oldAPIValue = this._getAPIValue();
    this._rawValue = val;
    this._dirtyValue = true;

    if (oldAPIValue !== this._getAPIValue()) {
      this._selectionStart = this._selectionEnd = this._getValueLength();
      this._selectionDirection = "none";
    ***REMOVED***
  ***REMOVED***

  get textLength() {
    return this.value.length; // code unit length (16 bit)
  ***REMOVED***

  get type() {
    return "textarea";
  ***REMOVED***

  _dispatchSelectEvent() {
    fireAnEvent("select", this, undefined, { bubbles: true, cancelable: true ***REMOVED***);
  ***REMOVED***

  _getValueLength() {
    return typeof this.value === "string" ? this.value.length : 0;
  ***REMOVED***

  select() {
    this._selectionStart = 0;
    this._selectionEnd = this._getValueLength();
    this._selectionDirection = "none";
    this._dispatchSelectEvent();
  ***REMOVED***

  get selectionStart() {
    return this._selectionStart;
  ***REMOVED***

  set selectionStart(start) {
    this.setSelectionRange(start, Math.max(start, this._selectionEnd), this._selectionDirection);
  ***REMOVED***

  get selectionEnd() {
    return this._selectionEnd;
  ***REMOVED***

  set selectionEnd(end) {
    this.setSelectionRange(this._selectionStart, end, this._selectionDirection);
  ***REMOVED***

  get selectionDirection() {
    return this._selectionDirection;
  ***REMOVED***

  set selectionDirection(dir) {
    this.setSelectionRange(this._selectionStart, this._selectionEnd, dir);
  ***REMOVED***

  setSelectionRange(start, end, dir) {
    this._selectionEnd = Math.min(end, this._getValueLength());
    this._selectionStart = Math.min(start, this._selectionEnd);
    this._selectionDirection = dir === "forward" || dir === "backward" ? dir : "none";
    this._dispatchSelectEvent();
  ***REMOVED***

  setRangeText(repl, start, end, selectionMode = "preserve") {
    if (arguments.length < 2) {
      start = this._selectionStart;
      end = this._selectionEnd;
    ***REMOVED*** else if (start > end) {
      throw DOMException.create(this._globalObject, ["The index is not in the allowed range.", "IndexSizeError"]);
    ***REMOVED***

    start = Math.min(start, this._getValueLength());
    end = Math.min(end, this._getValueLength());

    const val = this.value;
    let selStart = this._selectionStart;
    let selEnd = this._selectionEnd;

    this.value = val.slice(0, start) + repl + val.slice(end);

    const newEnd = start + this.value.length;

    if (selectionMode === "select") {
      this.setSelectionRange(start, newEnd);
    ***REMOVED*** else if (selectionMode === "start") {
      this.setSelectionRange(start, start);
    ***REMOVED*** else if (selectionMode === "end") {
      this.setSelectionRange(newEnd, newEnd);
    ***REMOVED*** else { // preserve
      const delta = repl.length - (end - start);

      if (selStart > end) {
        selStart += delta;
      ***REMOVED*** else if (selStart > start) {
        selStart = start;
      ***REMOVED***

      if (selEnd > end) {
        selEnd += delta;
      ***REMOVED*** else if (selEnd > start) {
        selEnd = newEnd;
      ***REMOVED***

      this.setSelectionRange(selStart, selEnd);
    ***REMOVED***
  ***REMOVED***

  get cols() {
    if (!this.hasAttributeNS(null, "cols")) {
      return 20;
    ***REMOVED***
    return parseInt(this.getAttributeNS(null, "cols"));
  ***REMOVED***

  set cols(value) {
    if (value <= 0) {
      throw DOMException.create(this._globalObject, ["The index is not in the allowed range.", "IndexSizeError"]);
    ***REMOVED***
    this.setAttributeNS(null, "cols", String(value));
  ***REMOVED***

  get rows() {
    if (!this.hasAttributeNS(null, "rows")) {
      return 2;
    ***REMOVED***
    return parseInt(this.getAttributeNS(null, "rows"));
  ***REMOVED***

  set rows(value) {
    if (value <= 0) {
      throw DOMException.create(this._globalObject, ["The index is not in the allowed range.", "IndexSizeError"]);
    ***REMOVED***
    this.setAttributeNS(null, "rows", String(value));
  ***REMOVED***

  _barredFromConstraintValidationSpecialization() {
    return this.hasAttributeNS(null, "readonly");
  ***REMOVED***

  get _mutable() {
    return !isDisabled(this) && !this.hasAttributeNS(null, "readonly");
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/form-elements.html#attr-textarea-required
  get validity() {
    if (!this._validity) {
      const state = {
        valueMissing: () => this.hasAttributeNS(null, "required") && this._mutable && this.value === ""
      ***REMOVED***;

      this._validity = ValidityState.createImpl(this._globalObject, [], {
        element: this,
        state
      ***REMOVED***);
    ***REMOVED***
    return this._validity;
  ***REMOVED***

  [cloningSteps](copy, node) {
    copy._dirtyValue = node._dirtyValue;
    copy._rawValue = node._rawValue;
  ***REMOVED***
***REMOVED***

mixin(HTMLTextAreaElementImpl.prototype, DefaultConstraintValidationImpl.prototype);

module.exports = {
  implementation: HTMLTextAreaElementImpl
***REMOVED***;

function textareaWrappingTransformation(text, cols) {
  let lineStart = 0;
  let lineEnd = text.indexOf("\n");
  if (lineEnd === -1) {
    lineEnd = text.length;
  ***REMOVED***

  while (lineStart < text.length) {
    const lineLength = lineEnd - lineStart;
    if (lineLength > cols) {
      // split the line
      lineEnd = lineStart + cols;
      text = text.slice(0, lineEnd) + "\n" + text.slice(lineEnd);
    ***REMOVED***
    // move to next line
    lineStart = lineEnd + 1; // step over the newline
    lineEnd = text.indexOf("\n", lineStart);
    if (lineEnd === -1) {
      lineEnd = text.length;
    ***REMOVED***
  ***REMOVED***

  return text;
***REMOVED***
