// core-js-compat/helpers.js
'use strict';
// eslint-disable-next-line es/no-object-hasown -- safe
const has = Object.hasOwn || Function.call.bind({***REMOVED***.hasOwnProperty);

const VERSION_PATTERN = /(\d+)(?:\.(\d+))?(?:\.(\d+))?/;

class SemVer {
  constructor(input) {
    const match = VERSION_PATTERN.exec(input);
    if (!match) throw new TypeError(`Invalid version: ${ input ***REMOVED***`);
    const [, $major, $minor, $patch] = match;
    this.major = +$major;
    this.minor = $minor ? +$minor : 0;
    this.patch = $patch ? +$patch : 0;
  ***REMOVED***
  toString() {
    return `${ this.major ***REMOVED***.${ this.minor ***REMOVED***.${ this.patch ***REMOVED***`;
  ***REMOVED***
***REMOVED***

function semver(input) {
  return input instanceof SemVer ? input : new SemVer(input);
***REMOVED***

function compare($a, operator, $b) {
  const a = semver($a);
  const b = semver($b);
  for (const component of ['major', 'minor', 'patch']) {
    if (a[component] < b[component]) return operator === '<' || operator === '<=' || operator === '!=';
    if (a[component] > b[component]) return operator === '>' || operator === '>=' || operator === '!=';
  ***REMOVED*** return operator === '==' || operator === '<=' || operator === '>=';
***REMOVED***

function filterOutStabilizedProposals(modules) {
  const modulesSet = new Set(modules);

  for (const $module of modulesSet) {
    if ($module.startsWith('esnext.') && modulesSet.has($module.replace(/^esnext\./, 'es.'))) {
      modulesSet.delete($module);
    ***REMOVED***
  ***REMOVED***

  return [...modulesSet];
***REMOVED***

function intersection(list, order) {
  const set = list instanceof Set ? list : new Set(list);
  return order.filter(name => set.has(name));
***REMOVED***

function sortObjectByKey(object, fn) {
  return Object.keys(object).sort(fn).reduce((memo, key) => {
    memo[key] = object[key];
    return memo;
  ***REMOVED***, {***REMOVED***);
***REMOVED***

module.exports = {
  compare,
  filterOutStabilizedProposals,
  has,
  intersection,
  semver,
  sortObjectByKey,
***REMOVED***;
