// @eslint/eslintrc/conf/environments.js
/**
 * @fileoverview Defines environment settings and globals.
 * @author Elan Shanker
 */

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

import globals from "globals";

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Get the object that has difference.
 * @param {Record<string,boolean>***REMOVED*** current The newer object.
 * @param {Record<string,boolean>***REMOVED*** prev The older object.
 * @returns {Record<string,boolean>***REMOVED*** The difference object.
 */
function getDiff(current, prev) {
    const retv = {***REMOVED***;

    for (const [key, value] of Object.entries(current)) {
        if (!Object.hasOwnProperty.call(prev, key)) {
            retv[key] = value;
        ***REMOVED***
    ***REMOVED***

    return retv;
***REMOVED***

const newGlobals2015 = getDiff(globals.es2015, globals.es5); // 19 variables such as Promise, Map, ...
const newGlobals2017 = {
    Atomics: false,
    SharedArrayBuffer: false
***REMOVED***;
const newGlobals2020 = {
    BigInt: false,
    BigInt64Array: false,
    BigUint64Array: false,
    globalThis: false
***REMOVED***;

const newGlobals2021 = {
    AggregateError: false,
    FinalizationRegistry: false,
    WeakRef: false
***REMOVED***;

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

/** @type {Map<string, import("../lib/shared/types").Environment>***REMOVED*** */
export default new Map(Object.entries({

    // Language
    builtin: {
        globals: globals.es5
    ***REMOVED***,
    es6: {
        globals: newGlobals2015,
        parserOptions: {
            ecmaVersion: 6
        ***REMOVED***
    ***REMOVED***,
    es2015: {
        globals: newGlobals2015,
        parserOptions: {
            ecmaVersion: 6
        ***REMOVED***
    ***REMOVED***,
    es2016: {
        globals: newGlobals2015,
        parserOptions: {
            ecmaVersion: 7
        ***REMOVED***
    ***REMOVED***,
    es2017: {
        globals: { ...newGlobals2015, ...newGlobals2017 ***REMOVED***,
        parserOptions: {
            ecmaVersion: 8
        ***REMOVED***
    ***REMOVED***,
    es2018: {
        globals: { ...newGlobals2015, ...newGlobals2017 ***REMOVED***,
        parserOptions: {
            ecmaVersion: 9
        ***REMOVED***
    ***REMOVED***,
    es2019: {
        globals: { ...newGlobals2015, ...newGlobals2017 ***REMOVED***,
        parserOptions: {
            ecmaVersion: 10
        ***REMOVED***
    ***REMOVED***,
    es2020: {
        globals: { ...newGlobals2015, ...newGlobals2017, ...newGlobals2020 ***REMOVED***,
        parserOptions: {
            ecmaVersion: 11
        ***REMOVED***
    ***REMOVED***,
    es2021: {
        globals: { ...newGlobals2015, ...newGlobals2017, ...newGlobals2020, ...newGlobals2021 ***REMOVED***,
        parserOptions: {
            ecmaVersion: 12
        ***REMOVED***
    ***REMOVED***,
    es2022: {
        globals: { ...newGlobals2015, ...newGlobals2017, ...newGlobals2020, ...newGlobals2021 ***REMOVED***,
        parserOptions: {
            ecmaVersion: 13
        ***REMOVED***
    ***REMOVED***,
    es2023: {
        globals: { ...newGlobals2015, ...newGlobals2017, ...newGlobals2020, ...newGlobals2021 ***REMOVED***,
        parserOptions: {
            ecmaVersion: 14
        ***REMOVED***
    ***REMOVED***,
    es2024: {
        globals: { ...newGlobals2015, ...newGlobals2017, ...newGlobals2020, ...newGlobals2021 ***REMOVED***,
        parserOptions: {
            ecmaVersion: 15
        ***REMOVED***
    ***REMOVED***,

    // Platforms
    browser: {
        globals: globals.browser
    ***REMOVED***,
    node: {
        globals: globals.node,
        parserOptions: {
            ecmaFeatures: {
                globalReturn: true
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***,
    "shared-node-browser": {
        globals: globals["shared-node-browser"]
    ***REMOVED***,
    worker: {
        globals: globals.worker
    ***REMOVED***,
    serviceworker: {
        globals: globals.serviceworker
    ***REMOVED***,

    // Frameworks
    commonjs: {
        globals: globals.commonjs,
        parserOptions: {
            ecmaFeatures: {
                globalReturn: true
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***,
    amd: {
        globals: globals.amd
    ***REMOVED***,
    mocha: {
        globals: globals.mocha
    ***REMOVED***,
    jasmine: {
        globals: globals.jasmine
    ***REMOVED***,
    jest: {
        globals: globals.jest
    ***REMOVED***,
    phantomjs: {
        globals: globals.phantomjs
    ***REMOVED***,
    jquery: {
        globals: globals.jquery
    ***REMOVED***,
    qunit: {
        globals: globals.qunit
    ***REMOVED***,
    prototypejs: {
        globals: globals.prototypejs
    ***REMOVED***,
    shelljs: {
        globals: globals.shelljs
    ***REMOVED***,
    meteor: {
        globals: globals.meteor
    ***REMOVED***,
    mongo: {
        globals: globals.mongo
    ***REMOVED***,
    protractor: {
        globals: globals.protractor
    ***REMOVED***,
    applescript: {
        globals: globals.applescript
    ***REMOVED***,
    nashorn: {
        globals: globals.nashorn
    ***REMOVED***,
    atomtest: {
        globals: globals.atomtest
    ***REMOVED***,
    embertest: {
        globals: globals.embertest
    ***REMOVED***,
    webextensions: {
        globals: globals.webextensions
    ***REMOVED***,
    greasemonkey: {
        globals: globals.greasemonkey
    ***REMOVED***
***REMOVED***));
