// qs/test/empty-keys-cases.js
'use strict';

module.exports = {
    emptyTestCases: [
        {
            input: '&',
            withEmptyKeys: {***REMOVED***,
            stringifyOutput: {
                brackets: '',
                indices: '',
                repeat: ''
            ***REMOVED***,
            noEmptyKeys: {***REMOVED***
        ***REMOVED***,
        {
            input: '&&',
            withEmptyKeys: {***REMOVED***,
            stringifyOutput: {
                brackets: '',
                indices: '',
                repeat: ''
            ***REMOVED***,
            noEmptyKeys: {***REMOVED***
        ***REMOVED***,
        {
            input: '&=',
            withEmptyKeys: { '': '' ***REMOVED***,
            stringifyOutput: {
                brackets: '=',
                indices: '=',
                repeat: '='
            ***REMOVED***,
            noEmptyKeys: {***REMOVED***
        ***REMOVED***,
        {
            input: '&=&',
            withEmptyKeys: { '': '' ***REMOVED***,
            stringifyOutput: {
                brackets: '=',
                indices: '=',
                repeat: '='
            ***REMOVED***,
            noEmptyKeys: {***REMOVED***
        ***REMOVED***,
        {
            input: '&=&=',
            withEmptyKeys: { '': ['', ''] ***REMOVED***,
            stringifyOutput: {
                brackets: '[]=&[]=',
                indices: '[0]=&[1]=',
                repeat: '=&='
            ***REMOVED***,
            noEmptyKeys: {***REMOVED***
        ***REMOVED***,
        {
            input: '&=&=&',
            withEmptyKeys: { '': ['', ''] ***REMOVED***,
            stringifyOutput: {
                brackets: '[]=&[]=',
                indices: '[0]=&[1]=',
                repeat: '=&='
            ***REMOVED***,
            noEmptyKeys: {***REMOVED***
        ***REMOVED***,
        {
            input: '=',
            withEmptyKeys: { '': '' ***REMOVED***,
            noEmptyKeys: {***REMOVED***,
            stringifyOutput: {
                brackets: '=',
                indices: '=',
                repeat: '='
            ***REMOVED***
        ***REMOVED***,
        {
            input: '=&',
            withEmptyKeys: { '': '' ***REMOVED***,
            stringifyOutput: {
                brackets: '=',
                indices: '=',
                repeat: '='
            ***REMOVED***,
            noEmptyKeys: {***REMOVED***
        ***REMOVED***,
        {
            input: '=&&&',
            withEmptyKeys: { '': '' ***REMOVED***,
            stringifyOutput: {
                brackets: '=',
                indices: '=',
                repeat: '='
            ***REMOVED***,
            noEmptyKeys: {***REMOVED***
        ***REMOVED***,
        {
            input: '=&=&=&',
            withEmptyKeys: { '': ['', '', ''] ***REMOVED***,
            stringifyOutput: {
                brackets: '[]=&[]=&[]=',
                indices: '[0]=&[1]=&[2]=',
                repeat: '=&=&='
            ***REMOVED***,
            noEmptyKeys: {***REMOVED***
        ***REMOVED***,
        {
            input: '=&a[]=b&a[1]=c',
            withEmptyKeys: { '': '', a: ['b', 'c'] ***REMOVED***,
            stringifyOutput: {
                brackets: '=&a[]=b&a[]=c',
                indices: '=&a[0]=b&a[1]=c',
                repeat: '=&a=b&a=c'
            ***REMOVED***,
            noEmptyKeys: { a: ['b', 'c'] ***REMOVED***
        ***REMOVED***,
        {
            input: '=a',
            withEmptyKeys: { '': 'a' ***REMOVED***,
            noEmptyKeys: {***REMOVED***,
            stringifyOutput: {
                brackets: '=a',
                indices: '=a',
                repeat: '=a'
            ***REMOVED***
        ***REMOVED***,
        {
            input: 'a==a',
            withEmptyKeys: { a: '=a' ***REMOVED***,
            noEmptyKeys: { a: '=a' ***REMOVED***,
            stringifyOutput: {
                brackets: 'a==a',
                indices: 'a==a',
                repeat: 'a==a'
            ***REMOVED***
        ***REMOVED***,
        {
            input: '=&a[]=b',
            withEmptyKeys: { '': '', a: ['b'] ***REMOVED***,
            stringifyOutput: {
                brackets: '=&a[]=b',
                indices: '=&a[0]=b',
                repeat: '=&a=b'
            ***REMOVED***,
            noEmptyKeys: { a: ['b'] ***REMOVED***
        ***REMOVED***,
        {
            input: '=&a[]=b&a[]=c&a[2]=d',
            withEmptyKeys: { '': '', a: ['b', 'c', 'd'] ***REMOVED***,
            stringifyOutput: {
                brackets: '=&a[]=b&a[]=c&a[]=d',
                indices: '=&a[0]=b&a[1]=c&a[2]=d',
                repeat: '=&a=b&a=c&a=d'
            ***REMOVED***,
            noEmptyKeys: { a: ['b', 'c', 'd'] ***REMOVED***
        ***REMOVED***,
        {
            input: '=a&=b',
            withEmptyKeys: { '': ['a', 'b'] ***REMOVED***,
            stringifyOutput: {
                brackets: '[]=a&[]=b',
                indices: '[0]=a&[1]=b',
                repeat: '=a&=b'
            ***REMOVED***,
            noEmptyKeys: {***REMOVED***
        ***REMOVED***,
        {
            input: '=a&foo=b',
            withEmptyKeys: { '': 'a', foo: 'b' ***REMOVED***,
            noEmptyKeys: { foo: 'b' ***REMOVED***,
            stringifyOutput: {
                brackets: '=a&foo=b',
                indices: '=a&foo=b',
                repeat: '=a&foo=b'
            ***REMOVED***
        ***REMOVED***,
        {
            input: 'a[]=b&a=c&=',
            withEmptyKeys: { '': '', a: ['b', 'c'] ***REMOVED***,
            stringifyOutput: {
                brackets: '=&a[]=b&a[]=c',
                indices: '=&a[0]=b&a[1]=c',
                repeat: '=&a=b&a=c'
            ***REMOVED***,
            noEmptyKeys: { a: ['b', 'c'] ***REMOVED***
        ***REMOVED***,
        {
            input: 'a[]=b&a=c&=',
            withEmptyKeys: { '': '', a: ['b', 'c'] ***REMOVED***,
            stringifyOutput: {
                brackets: '=&a[]=b&a[]=c',
                indices: '=&a[0]=b&a[1]=c',
                repeat: '=&a=b&a=c'
            ***REMOVED***,
            noEmptyKeys: { a: ['b', 'c'] ***REMOVED***
        ***REMOVED***,
        {
            input: 'a[0]=b&a=c&=',
            withEmptyKeys: { '': '', a: ['b', 'c'] ***REMOVED***,
            stringifyOutput: {
                brackets: '=&a[]=b&a[]=c',
                indices: '=&a[0]=b&a[1]=c',
                repeat: '=&a=b&a=c'
            ***REMOVED***,
            noEmptyKeys: { a: ['b', 'c'] ***REMOVED***
        ***REMOVED***,
        {
            input: 'a=b&a[]=c&=',
            withEmptyKeys: { '': '', a: ['b', 'c'] ***REMOVED***,
            stringifyOutput: {
                brackets: '=&a[]=b&a[]=c',
                indices: '=&a[0]=b&a[1]=c',
                repeat: '=&a=b&a=c'
            ***REMOVED***,
            noEmptyKeys: { a: ['b', 'c'] ***REMOVED***
        ***REMOVED***,
        {
            input: 'a=b&a[0]=c&=',
            withEmptyKeys: { '': '', a: ['b', 'c'] ***REMOVED***,
            stringifyOutput: {
                brackets: '=&a[]=b&a[]=c',
                indices: '=&a[0]=b&a[1]=c',
                repeat: '=&a=b&a=c'
            ***REMOVED***,
            noEmptyKeys: { a: ['b', 'c'] ***REMOVED***
        ***REMOVED***,
        {
            input: '[]=a&[]=b& []=1',
            withEmptyKeys: { '': ['a', 'b'], ' ': ['1'] ***REMOVED***,
            stringifyOutput: {
                brackets: '[]=a&[]=b& []=1',
                indices: '[0]=a&[1]=b& [0]=1',
                repeat: '=a&=b& =1'
            ***REMOVED***,
            noEmptyKeys: { 0: 'a', 1: 'b', ' ': ['1'] ***REMOVED***
        ***REMOVED***,
        {
            input: '[0]=a&[1]=b&a[0]=1&a[1]=2',
            withEmptyKeys: { '': ['a', 'b'], a: ['1', '2'] ***REMOVED***,
            noEmptyKeys: { 0: 'a', 1: 'b', a: ['1', '2'] ***REMOVED***,
            stringifyOutput: {
                brackets: '[]=a&[]=b&a[]=1&a[]=2',
                indices: '[0]=a&[1]=b&a[0]=1&a[1]=2',
                repeat: '=a&=b&a=1&a=2'
            ***REMOVED***
        ***REMOVED***,
        {
            input: '[deep]=a&[deep]=2',
            withEmptyKeys: { '': { deep: ['a', '2'] ***REMOVED***
            ***REMOVED***,
            stringifyOutput: {
                brackets: '[deep][]=a&[deep][]=2',
                indices: '[deep][0]=a&[deep][1]=2',
                repeat: '[deep]=a&[deep]=2'
            ***REMOVED***,
            noEmptyKeys: { deep: ['a', '2'] ***REMOVED***
        ***REMOVED***,
        {
            input: '%5B0%5D=a&%5B1%5D=b',
            withEmptyKeys: { '': ['a', 'b'] ***REMOVED***,
            stringifyOutput: {
                brackets: '[]=a&[]=b',
                indices: '[0]=a&[1]=b',
                repeat: '=a&=b'
            ***REMOVED***,
            noEmptyKeys: { 0: 'a', 1: 'b' ***REMOVED***
        ***REMOVED***
    ]
***REMOVED***;
