// jsdom/lib/jsdom/living/events/Event-impl.js
"use strict";

const idlUtils = require("../generated/utils");
const EventInit = require("../generated/EventInit");

class EventImpl {
  constructor(globalObject, args, privateData) {
    const [type, eventInitDict = this.constructor.defaultInit] = args;

    this.type = type;

    this.bubbles = false;
    this.cancelable = false;
    for (const key in eventInitDict) {
      if (key in this.constructor.defaultInit) {
        this[key] = eventInitDict[key];
      ***REMOVED***
    ***REMOVED***
    for (const key in this.constructor.defaultInit) {
      if (!(key in this)) {
        this[key] = this.constructor.defaultInit[key];
      ***REMOVED***
    ***REMOVED***

    this.target = null;
    this.currentTarget = null;
    this.eventPhase = 0;

    this._globalObject = globalObject;
    this._initializedFlag = true;
    this._stopPropagationFlag = false;
    this._stopImmediatePropagationFlag = false;
    this._canceledFlag = false;
    this._inPassiveListenerFlag = false;
    this._dispatchFlag = false;
    this._path = [];

    this.isTrusted = privateData.isTrusted || false;
    this.timeStamp = Date.now();
  ***REMOVED***

  // https://dom.spec.whatwg.org/#set-the-canceled-flag
  _setTheCanceledFlag() {
    if (this.cancelable && !this._inPassiveListenerFlag) {
      this._canceledFlag = true;
    ***REMOVED***
  ***REMOVED***

  get srcElement() {
    return this.target;
  ***REMOVED***

  get returnValue() {
    return !this._canceledFlag;
  ***REMOVED***

  set returnValue(v) {
    if (v === false) {
      this._setTheCanceledFlag();
    ***REMOVED***
  ***REMOVED***

  get defaultPrevented() {
    return this._canceledFlag;
  ***REMOVED***

  stopPropagation() {
    this._stopPropagationFlag = true;
  ***REMOVED***

  get cancelBubble() {
    return this._stopPropagationFlag;
  ***REMOVED***

  set cancelBubble(v) {
    if (v) {
      this._stopPropagationFlag = true;
    ***REMOVED***
  ***REMOVED***

  stopImmediatePropagation() {
    this._stopPropagationFlag = true;
    this._stopImmediatePropagationFlag = true;
  ***REMOVED***

  preventDefault() {
    this._setTheCanceledFlag();
  ***REMOVED***

  // https://dom.spec.whatwg.org/#dom-event-composedpath
  // Current implementation is based of https://whatpr.org/dom/699.html#dom-event-composedpath
  // due to a bug in composed path implementation https://github.com/whatwg/dom/issues/684
  composedPath() {
    const composedPath = [];

    const { currentTarget, _path: path ***REMOVED*** = this;

    if (path.length === 0) {
      return composedPath;
    ***REMOVED***

    composedPath.push(currentTarget);

    let currentTargetIndex = 0;
    let currentTargetHiddenSubtreeLevel = 0;

    for (let index = path.length - 1; index >= 0; index--) {
      const { item, rootOfClosedTree, slotInClosedTree ***REMOVED*** = path[index];

      if (rootOfClosedTree) {
        currentTargetHiddenSubtreeLevel++;
      ***REMOVED***

      if (item === idlUtils.implForWrapper(currentTarget)) {
        currentTargetIndex = index;
        break;
      ***REMOVED***

      if (slotInClosedTree) {
        currentTargetHiddenSubtreeLevel--;
      ***REMOVED***
    ***REMOVED***

    let currentHiddenLevel = currentTargetHiddenSubtreeLevel;
    let maxHiddenLevel = currentTargetHiddenSubtreeLevel;

    for (let i = currentTargetIndex - 1; i >= 0; i--) {
      const { item, rootOfClosedTree, slotInClosedTree ***REMOVED*** = path[i];

      if (rootOfClosedTree) {
        currentHiddenLevel++;
      ***REMOVED***

      if (currentHiddenLevel <= maxHiddenLevel) {
        composedPath.unshift(idlUtils.wrapperForImpl(item));
      ***REMOVED***

      if (slotInClosedTree) {
        currentHiddenLevel--;
        if (currentHiddenLevel < maxHiddenLevel) {
          maxHiddenLevel = currentHiddenLevel;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    currentHiddenLevel = currentTargetHiddenSubtreeLevel;
    maxHiddenLevel = currentTargetHiddenSubtreeLevel;

    for (let index = currentTargetIndex + 1; index < path.length; index++) {
      const { item, rootOfClosedTree, slotInClosedTree ***REMOVED*** = path[index];

      if (slotInClosedTree) {
        currentHiddenLevel++;
      ***REMOVED***

      if (currentHiddenLevel <= maxHiddenLevel) {
        composedPath.push(idlUtils.wrapperForImpl(item));
      ***REMOVED***

      if (rootOfClosedTree) {
        currentHiddenLevel--;
        if (currentHiddenLevel < maxHiddenLevel) {
          maxHiddenLevel = currentHiddenLevel;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    return composedPath;
  ***REMOVED***

  _initialize(type, bubbles, cancelable) {
    this.type = type;
    this._initializedFlag = true;

    this._stopPropagationFlag = false;
    this._stopImmediatePropagationFlag = false;
    this._canceledFlag = false;

    this.isTrusted = false;
    this.target = null;
    this.bubbles = bubbles;
    this.cancelable = cancelable;
  ***REMOVED***

  initEvent(type, bubbles, cancelable) {
    if (this._dispatchFlag) {
      return;
    ***REMOVED***

    this._initialize(type, bubbles, cancelable);
  ***REMOVED***
***REMOVED***
EventImpl.defaultInit = EventInit.convert(undefined);

module.exports = {
  implementation: EventImpl
***REMOVED***;
