// postcss-discard-empty/src/index.js
'use strict';
const plugin = 'postcss-discard-empty';
/**
 * @param {import('postcss').Root***REMOVED*** css
 * @param {import('postcss').Result***REMOVED*** result
 * @return {void***REMOVED***
 */
function discardAndReport(css, result) {
  /**
   * @param {import('postcss').AnyNode***REMOVED*** node
   * @return {void***REMOVED***
   */
  function discardEmpty(node) {
    const { type ***REMOVED*** = node;
    /** @type {(import('postcss').ChildNode | import('postcss').ChildProps)[] | undefined***REMOVED*** */
    const sub = /** @type {any***REMOVED*** */ (node).nodes;
    if (sub) {
      /** @type {import('postcss').Container***REMOVED*** */ (node).each(discardEmpty);
    ***REMOVED***

    if (
      (type === 'decl' && !node.value && !node.prop.startsWith('--')) ||
      (type === 'rule' && !node.selector) ||
      (sub && !sub.length) ||
      (type === 'atrule' &&
        ((!sub && !node.params) ||
          (!node.params &&
            !(/** @type {import('postcss').ChildNode[]***REMOVED****/ (sub).length))))
    ) {
      node.remove();

      result.messages.push({
        type: 'removal',
        plugin,
        node,
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***

  css.each(discardEmpty);
***REMOVED***

/**
 * @type {import('postcss').PluginCreator<void>***REMOVED***
 * @return {import('postcss').Plugin***REMOVED***
 */
function pluginCreator() {
  return {
    postcssPlugin: plugin,
    OnceExit(css, { result ***REMOVED***) {
      discardAndReport(css, result);
    ***REMOVED***,
  ***REMOVED***;
***REMOVED***

pluginCreator.postcss = true;
module.exports = pluginCreator;
