// ajv/dist/vocabularies/validation/pattern.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const code_1 = require("../code");
const codegen_1 = require("../../compile/codegen");
const error = {
    message: ({ schemaCode ***REMOVED***) => (0, codegen_1.str) `must match pattern "${schemaCode***REMOVED***"`,
    params: ({ schemaCode ***REMOVED***) => (0, codegen_1._) `{pattern: ${schemaCode***REMOVED******REMOVED***`,
***REMOVED***;
const def = {
    keyword: "pattern",
    type: "string",
    schemaType: "string",
    $data: true,
    error,
    code(cxt) {
        const { data, $data, schema, schemaCode, it ***REMOVED*** = cxt;
        // TODO regexp should be wrapped in try/catchs
        const u = it.opts.unicodeRegExp ? "u" : "";
        const regExp = $data ? (0, codegen_1._) `(new RegExp(${schemaCode***REMOVED***, ${u***REMOVED***))` : (0, code_1.usePattern)(cxt, schema);
        cxt.fail$data((0, codegen_1._) `!${regExp***REMOVED***.test(${data***REMOVED***)`);
    ***REMOVED***,
***REMOVED***;
exports.default = def;
