// ajv/dist/vocabularies/validation/required.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const code_1 = require("../code");
const codegen_1 = require("../../compile/codegen");
const util_1 = require("../../compile/util");
const error = {
    message: ({ params: { missingProperty ***REMOVED*** ***REMOVED***) => (0, codegen_1.str) `must have required property '${missingProperty***REMOVED***'`,
    params: ({ params: { missingProperty ***REMOVED*** ***REMOVED***) => (0, codegen_1._) `{missingProperty: ${missingProperty***REMOVED******REMOVED***`,
***REMOVED***;
const def = {
    keyword: "required",
    type: "object",
    schemaType: "array",
    $data: true,
    error,
    code(cxt) {
        const { gen, schema, schemaCode, data, $data, it ***REMOVED*** = cxt;
        const { opts ***REMOVED*** = it;
        if (!$data && schema.length === 0)
            return;
        const useLoop = schema.length >= opts.loopRequired;
        if (it.allErrors)
            allErrorsMode();
        else
            exitOnErrorMode();
        if (opts.strictRequired) {
            const props = cxt.parentSchema.properties;
            const { definedProperties ***REMOVED*** = cxt.it;
            for (const requiredKey of schema) {
                if ((props === null || props === void 0 ? void 0 : props[requiredKey]) === undefined && !definedProperties.has(requiredKey)) {
                    const schemaPath = it.schemaEnv.baseId + it.errSchemaPath;
                    const msg = `required property "${requiredKey***REMOVED***" is not defined at "${schemaPath***REMOVED***" (strictRequired)`;
                    (0, util_1.checkStrictMode)(it, msg, it.opts.strictRequired);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        function allErrorsMode() {
            if (useLoop || $data) {
                cxt.block$data(codegen_1.nil, loopAllRequired);
            ***REMOVED***
            else {
                for (const prop of schema) {
                    (0, code_1.checkReportMissingProp)(cxt, prop);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        function exitOnErrorMode() {
            const missing = gen.let("missing");
            if (useLoop || $data) {
                const valid = gen.let("valid", true);
                cxt.block$data(valid, () => loopUntilMissing(missing, valid));
                cxt.ok(valid);
            ***REMOVED***
            else {
                gen.if((0, code_1.checkMissingProp)(cxt, schema, missing));
                (0, code_1.reportMissingProp)(cxt, missing);
                gen.else();
            ***REMOVED***
        ***REMOVED***
        function loopAllRequired() {
            gen.forOf("prop", schemaCode, (prop) => {
                cxt.setParams({ missingProperty: prop ***REMOVED***);
                gen.if((0, code_1.noPropertyInData)(gen, data, prop, opts.ownProperties), () => cxt.error());
            ***REMOVED***);
        ***REMOVED***
        function loopUntilMissing(missing, valid) {
            cxt.setParams({ missingProperty: missing ***REMOVED***);
            gen.forOf(missing, schemaCode, () => {
                gen.assign(valid, (0, code_1.propertyInData)(gen, data, missing, opts.ownProperties));
                gen.if((0, codegen_1.not)(valid), () => {
                    cxt.error();
                    gen.break();
                ***REMOVED***);
            ***REMOVED***, codegen_1.nil);
        ***REMOVED***
    ***REMOVED***,
***REMOVED***;
exports.default = def;
