// enhanced-resolve/lib/AliasFieldPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const DescriptionFileUtils = require("./DescriptionFileUtils");
const getInnerRequest = require("./getInnerRequest");

/** @typedef {import("./Resolver")***REMOVED*** Resolver */
/** @typedef {import("./Resolver").JsonPrimitive***REMOVED*** JsonPrimitive */
/** @typedef {import("./Resolver").ResolveRequest***REMOVED*** ResolveRequest */
/** @typedef {import("./Resolver").ResolveStepHook***REMOVED*** ResolveStepHook */

module.exports = class AliasFieldPlugin {
	/**
	 * @param {string | ResolveStepHook***REMOVED*** source source
	 * @param {string | Array<string>***REMOVED*** field field
	 * @param {string | ResolveStepHook***REMOVED*** target target
	 */
	constructor(source, field, target) {
		this.source = source;
		this.field = field;
		this.target = target;
	***REMOVED***

	/**
	 * @param {Resolver***REMOVED*** resolver the resolver
	 * @returns {void***REMOVED***
	 */
	apply(resolver) {
		const target = resolver.ensureHook(this.target);
		resolver
			.getHook(this.source)
			.tapAsync("AliasFieldPlugin", (request, resolveContext, callback) => {
				if (!request.descriptionFileData) return callback();
				const innerRequest = getInnerRequest(resolver, request);
				if (!innerRequest) return callback();
				const fieldData = DescriptionFileUtils.getField(
					request.descriptionFileData,
					this.field
				);
				if (fieldData === null || typeof fieldData !== "object") {
					if (resolveContext.log)
						resolveContext.log(
							"Field '" +
								this.field +
								"' doesn't contain a valid alias configuration"
						);
					return callback();
				***REMOVED***
				/** @type {JsonPrimitive | undefined***REMOVED*** */
				const data = Object.prototype.hasOwnProperty.call(
					fieldData,
					innerRequest
				)
					? /** @type {{[Key in string]: JsonPrimitive***REMOVED******REMOVED*** */ (fieldData)[
							innerRequest
					  ]
					: innerRequest.startsWith("./")
					? /** @type {{[Key in string]: JsonPrimitive***REMOVED******REMOVED*** */ (fieldData)[
							innerRequest.slice(2)
					  ]
					: undefined;
				if (data === innerRequest) return callback();
				if (data === undefined) return callback();
				if (data === false) {
					/** @type {ResolveRequest***REMOVED*** */
					const ignoreObj = {
						...request,
						path: false
					***REMOVED***;
					if (typeof resolveContext.yield === "function") {
						resolveContext.yield(ignoreObj);
						return callback(null, null);
					***REMOVED***
					return callback(null, ignoreObj);
				***REMOVED***
				/** @type {ResolveRequest***REMOVED*** */
				const obj = {
					...request,
					path: /** @type {string***REMOVED*** */ (request.descriptionFileRoot),
					request: /** @type {string***REMOVED*** */ (data),
					fullySpecified: false
				***REMOVED***;
				resolver.doResolve(
					target,
					obj,
					"aliased from description file " +
						request.descriptionFilePath +
						" with mapping '" +
						innerRequest +
						"' to '" +
						/** @type {string***REMOVED*** */ (data) +
						"'",
					resolveContext,
					(err, result) => {
						if (err) return callback(err);

						// Don't allow other aliasing or raw request
						if (result === undefined) return callback(null, null);
						callback(null, result);
					***REMOVED***
				);
			***REMOVED***);
	***REMOVED***
***REMOVED***;
