// enhanced-resolve/lib/AliasPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const forEachBail = require("./forEachBail");
const { PathType, getType ***REMOVED*** = require("./util/path");

/** @typedef {import("./Resolver")***REMOVED*** Resolver */
/** @typedef {import("./Resolver").ResolveRequest***REMOVED*** ResolveRequest */
/** @typedef {import("./Resolver").ResolveStepHook***REMOVED*** ResolveStepHook */
/** @typedef {string | Array<string> | false***REMOVED*** Alias */
/** @typedef {{alias: Alias, name: string, onlyModule?: boolean***REMOVED******REMOVED*** AliasOption */

module.exports = class AliasPlugin {
	/**
	 * @param {string | ResolveStepHook***REMOVED*** source source
	 * @param {AliasOption | Array<AliasOption>***REMOVED*** options options
	 * @param {string | ResolveStepHook***REMOVED*** target target
	 */
	constructor(source, options, target) {
		this.source = source;
		this.options = Array.isArray(options) ? options : [options];
		this.target = target;
	***REMOVED***

	/**
	 * @param {Resolver***REMOVED*** resolver the resolver
	 * @returns {void***REMOVED***
	 */
	apply(resolver) {
		const target = resolver.ensureHook(this.target);
		/**
		 * @param {string***REMOVED*** maybeAbsolutePath path
		 * @returns {null|string***REMOVED*** absolute path with slash ending
		 */
		const getAbsolutePathWithSlashEnding = maybeAbsolutePath => {
			const type = getType(maybeAbsolutePath);
			if (type === PathType.AbsolutePosix || type === PathType.AbsoluteWin) {
				return resolver.join(maybeAbsolutePath, "_").slice(0, -1);
			***REMOVED***
			return null;
		***REMOVED***;
		/**
		 * @param {string***REMOVED*** path path
		 * @param {string***REMOVED*** maybeSubPath sub path
		 * @returns {boolean***REMOVED*** true, if path is sub path
		 */
		const isSubPath = (path, maybeSubPath) => {
			const absolutePath = getAbsolutePathWithSlashEnding(maybeSubPath);
			if (!absolutePath) return false;
			return path.startsWith(absolutePath);
		***REMOVED***;
		resolver
			.getHook(this.source)
			.tapAsync("AliasPlugin", (request, resolveContext, callback) => {
				const innerRequest = request.request || request.path;
				if (!innerRequest) return callback();

				forEachBail(
					this.options,
					(item, callback) => {
						/** @type {boolean***REMOVED*** */
						let shouldStop = false;

						const matchRequest =
							innerRequest === item.name ||
							(!item.onlyModule &&
								(request.request
									? innerRequest.startsWith(`${item.name***REMOVED***/`)
									: isSubPath(innerRequest, item.name)));

						const splitName = item.name.split("*");
						const matchWildcard = !item.onlyModule && splitName.length === 2;

						if (matchRequest || matchWildcard) {
							/**
							 * @param {Alias***REMOVED*** alias alias
							 * @param {(err?: null|Error, result?: null|ResolveRequest) => void***REMOVED*** callback callback
							 * @returns {void***REMOVED***
							 */
							const resolveWithAlias = (alias, callback) => {
								if (alias === false) {
									/** @type {ResolveRequest***REMOVED*** */
									const ignoreObj = {
										...request,
										path: false
									***REMOVED***;
									if (typeof resolveContext.yield === "function") {
										resolveContext.yield(ignoreObj);
										return callback(null, null);
									***REMOVED***
									return callback(null, ignoreObj);
								***REMOVED***

								let newRequestStr;

								const [prefix, suffix] = splitName;
								if (
									matchWildcard &&
									innerRequest.startsWith(prefix) &&
									innerRequest.endsWith(suffix)
								) {
									const match = innerRequest.slice(
										prefix.length,
										innerRequest.length - suffix.length
									);
									newRequestStr = item.alias.toString().replace("*", match);
								***REMOVED***

								if (
									matchRequest &&
									innerRequest !== alias &&
									!innerRequest.startsWith(alias + "/")
								) {
									/** @type {string***REMOVED*** */
									const remainingRequest = innerRequest.slice(item.name.length);
									newRequestStr = alias + remainingRequest;
								***REMOVED***

								if (newRequestStr !== undefined) {
									shouldStop = true;
									/** @type {ResolveRequest***REMOVED*** */
									const obj = {
										...request,
										request: newRequestStr,
										fullySpecified: false
									***REMOVED***;
									return resolver.doResolve(
										target,
										obj,
										"aliased with mapping '" +
											item.name +
											"': '" +
											alias +
											"' to '" +
											newRequestStr +
											"'",
										resolveContext,
										(err, result) => {
											if (err) return callback(err);
											if (result) return callback(null, result);
											return callback();
										***REMOVED***
									);
								***REMOVED***
								return callback();
							***REMOVED***;

							/**
							 * @param {null|Error***REMOVED*** [err] error
							 * @param {null|ResolveRequest***REMOVED*** [result] result
							 * @returns {void***REMOVED***
							 */
							const stoppingCallback = (err, result) => {
								if (err) return callback(err);

								if (result) return callback(null, result);
								// Don't allow other aliasing or raw request
								if (shouldStop) return callback(null, null);
								return callback();
							***REMOVED***;

							if (Array.isArray(item.alias)) {
								return forEachBail(
									item.alias,
									resolveWithAlias,
									stoppingCallback
								);
							***REMOVED*** else {
								return resolveWithAlias(item.alias, stoppingCallback);
							***REMOVED***
						***REMOVED***

						return callback();
					***REMOVED***,
					callback
				);
			***REMOVED***);
	***REMOVED***
***REMOVED***;
