// enhanced-resolve/lib/CachedInputFileSystem.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const nextTick = require("process").nextTick;

/** @typedef {import("./Resolver").FileSystem***REMOVED*** FileSystem */
/** @typedef {import("./Resolver").PathLike***REMOVED*** PathLike */
/** @typedef {import("./Resolver").PathOrFileDescriptor***REMOVED*** PathOrFileDescriptor */
/** @typedef {import("./Resolver").SyncFileSystem***REMOVED*** SyncFileSystem */
/** @typedef {FileSystem & SyncFileSystem***REMOVED*** BaseFileSystem */

/**
 * @template T
 * @typedef {import("./Resolver").FileSystemCallback<T>***REMOVED*** FileSystemCallback<T>
 */

/**
 * @param {string***REMOVED*** path path
 * @returns {string***REMOVED*** dirname
 */
const dirname = path => {
	let idx = path.length - 1;
	while (idx >= 0) {
		const c = path.charCodeAt(idx);
		// slash or backslash
		if (c === 47 || c === 92) break;
		idx--;
	***REMOVED***
	if (idx < 0) return "";
	return path.slice(0, idx);
***REMOVED***;

/**
 * @template T
 * @param {FileSystemCallback<T>[]***REMOVED*** callbacks callbacks
 * @param {Error | null***REMOVED*** err error
 * @param {T***REMOVED*** result result
 */
const runCallbacks = (callbacks, err, result) => {
	if (callbacks.length === 1) {
		callbacks[0](err, result);
		callbacks.length = 0;
		return;
	***REMOVED***
	let error;
	for (const callback of callbacks) {
		try {
			callback(err, result);
		***REMOVED*** catch (e) {
			if (!error) error = e;
		***REMOVED***
	***REMOVED***
	callbacks.length = 0;
	if (error) throw error;
***REMOVED***;

class OperationMergerBackend {
	/**
	 * @param {Function | undefined***REMOVED*** provider async method in filesystem
	 * @param {Function | undefined***REMOVED*** syncProvider sync method in filesystem
	 * @param {BaseFileSystem***REMOVED*** providerContext call context for the provider methods
	 */
	constructor(provider, syncProvider, providerContext) {
		this._provider = provider;
		this._syncProvider = syncProvider;
		this._providerContext = providerContext;
		this._activeAsyncOperations = new Map();

		this.provide = this._provider
			? /**
			   * @param {PathLike | PathOrFileDescriptor***REMOVED*** path path
			   * @param {object | FileSystemCallback<any> | undefined***REMOVED*** options options
			   * @param {FileSystemCallback<any>=***REMOVED*** callback callback
			   * @returns {any***REMOVED*** result
			   */
			  (path, options, callback) => {
					if (typeof options === "function") {
						callback = /** @type {FileSystemCallback<any>***REMOVED*** */ (options);
						options = undefined;
					***REMOVED***
					if (
						typeof path !== "string" &&
						!Buffer.isBuffer(path) &&
						!(path instanceof URL) &&
						typeof path !== "number"
					) {
						/** @type {Function***REMOVED*** */
						(callback)(
							new TypeError("path must be a string, Buffer, URL or number")
						);
						return;
					***REMOVED***
					if (options) {
						return /** @type {Function***REMOVED*** */ (this._provider).call(
							this._providerContext,
							path,
							options,
							callback
						);
					***REMOVED***
					let callbacks = this._activeAsyncOperations.get(path);
					if (callbacks) {
						callbacks.push(callback);
						return;
					***REMOVED***
					this._activeAsyncOperations.set(path, (callbacks = [callback]));
					/** @type {Function***REMOVED*** */
					(provider)(
						path,
						/**
						 * @param {Error***REMOVED*** err error
						 * @param {any***REMOVED*** result result
						 */
						(err, result) => {
							this._activeAsyncOperations.delete(path);
							runCallbacks(callbacks, err, result);
						***REMOVED***
					);
			  ***REMOVED***
			: null;
		this.provideSync = this._syncProvider
			? /**
			   * @param {PathLike | PathOrFileDescriptor***REMOVED*** path path
			   * @param {object=***REMOVED*** options options
			   * @returns {any***REMOVED*** result
			   */
			  (path, options) => {
					return /** @type {Function***REMOVED*** */ (this._syncProvider).call(
						this._providerContext,
						path,
						options
					);
			  ***REMOVED***
			: null;
	***REMOVED***

	purge() {***REMOVED***
	purgeParent() {***REMOVED***
***REMOVED***

/*

IDLE:
	insert data: goto SYNC

SYNC:
	before provide: run ticks
	event loop tick: goto ASYNC_ACTIVE

ASYNC:
	timeout: run tick, goto ASYNC_PASSIVE

ASYNC_PASSIVE:
	before provide: run ticks

IDLE --[insert data]--> SYNC --[event loop tick]--> ASYNC_ACTIVE --[interval tick]-> ASYNC_PASSIVE
                                                          ^                             |
                                                          +---------[insert data]-------+
*/

const STORAGE_MODE_IDLE = 0;
const STORAGE_MODE_SYNC = 1;
const STORAGE_MODE_ASYNC = 2;

/**
 * @callback Provide
 * @param {PathLike | PathOrFileDescriptor***REMOVED*** path path
 * @param {any***REMOVED*** options options
 * @param {FileSystemCallback<any>***REMOVED*** callback callback
 * @returns {void***REMOVED***
 */

class CacheBackend {
	/**
	 * @param {number***REMOVED*** duration max cache duration of items
	 * @param {function | undefined***REMOVED*** provider async method
	 * @param {function | undefined***REMOVED*** syncProvider sync method
	 * @param {BaseFileSystem***REMOVED*** providerContext call context for the provider methods
	 */
	constructor(duration, provider, syncProvider, providerContext) {
		this._duration = duration;
		this._provider = provider;
		this._syncProvider = syncProvider;
		this._providerContext = providerContext;
		/** @type {Map<string, FileSystemCallback<any>[]>***REMOVED*** */
		this._activeAsyncOperations = new Map();
		/** @type {Map<string, { err: Error | null, result?: any, level: Set<string> ***REMOVED***>***REMOVED*** */
		this._data = new Map();
		/** @type {Set<string>[]***REMOVED*** */
		this._levels = [];
		for (let i = 0; i < 10; i++) this._levels.push(new Set());
		for (let i = 5000; i < duration; i += 500) this._levels.push(new Set());
		this._currentLevel = 0;
		this._tickInterval = Math.floor(duration / this._levels.length);
		/** @type {STORAGE_MODE_IDLE | STORAGE_MODE_SYNC | STORAGE_MODE_ASYNC***REMOVED*** */
		this._mode = STORAGE_MODE_IDLE;

		/** @type {NodeJS.Timeout | undefined***REMOVED*** */
		this._timeout = undefined;
		/** @type {number | undefined***REMOVED*** */
		this._nextDecay = undefined;

		// @ts-ignore
		this.provide = provider ? this.provide.bind(this) : null;
		// @ts-ignore
		this.provideSync = syncProvider ? this.provideSync.bind(this) : null;
	***REMOVED***

	/**
	 * @param {PathLike | PathOrFileDescriptor***REMOVED*** path path
	 * @param {any***REMOVED*** options options
	 * @param {FileSystemCallback<any>***REMOVED*** callback callback
	 * @returns {void***REMOVED***
	 */
	provide(path, options, callback) {
		if (typeof options === "function") {
			callback = options;
			options = undefined;
		***REMOVED***
		if (
			typeof path !== "string" &&
			!Buffer.isBuffer(path) &&
			!(path instanceof URL) &&
			typeof path !== "number"
		) {
			callback(new TypeError("path must be a string, Buffer, URL or number"));
			return;
		***REMOVED***
		const strPath = typeof path !== "string" ? path.toString() : path;
		if (options) {
			return /** @type {Function***REMOVED*** */ (this._provider).call(
				this._providerContext,
				path,
				options,
				callback
			);
		***REMOVED***

		// When in sync mode we can move to async mode
		if (this._mode === STORAGE_MODE_SYNC) {
			this._enterAsyncMode();
		***REMOVED***

		// Check in cache
		let cacheEntry = this._data.get(strPath);
		if (cacheEntry !== undefined) {
			if (cacheEntry.err) return nextTick(callback, cacheEntry.err);
			return nextTick(callback, null, cacheEntry.result);
		***REMOVED***

		// Check if there is already the same operation running
		let callbacks = this._activeAsyncOperations.get(strPath);
		if (callbacks !== undefined) {
			callbacks.push(callback);
			return;
		***REMOVED***
		this._activeAsyncOperations.set(strPath, (callbacks = [callback]));

		// Run the operation
		/** @type {Function***REMOVED*** */
		(this._provider).call(
			this._providerContext,
			path,
			/**
			 * @param {Error | null***REMOVED*** err error
			 * @param {any***REMOVED*** [result] result
			 */
			(err, result) => {
				this._activeAsyncOperations.delete(strPath);
				this._storeResult(strPath, err, result);

				// Enter async mode if not yet done
				this._enterAsyncMode();

				runCallbacks(
					/** @type {FileSystemCallback<any>[]***REMOVED*** */ (callbacks),
					err,
					result
				);
			***REMOVED***
		);
	***REMOVED***

	/**
	 * @param {PathLike | PathOrFileDescriptor***REMOVED*** path path
	 * @param {any***REMOVED*** options options
	 * @returns {any***REMOVED*** result
	 */
	provideSync(path, options) {
		if (
			typeof path !== "string" &&
			!Buffer.isBuffer(path) &&
			!(path instanceof URL) &&
			typeof path !== "number"
		) {
			throw new TypeError("path must be a string");
		***REMOVED***
		const strPath = typeof path !== "string" ? path.toString() : path;
		if (options) {
			return /** @type {Function***REMOVED*** */ (this._syncProvider).call(
				this._providerContext,
				path,
				options
			);
		***REMOVED***

		// In sync mode we may have to decay some cache items
		if (this._mode === STORAGE_MODE_SYNC) {
			this._runDecays();
		***REMOVED***

		// Check in cache
		let cacheEntry = this._data.get(strPath);
		if (cacheEntry !== undefined) {
			if (cacheEntry.err) throw cacheEntry.err;
			return cacheEntry.result;
		***REMOVED***

		// Get all active async operations
		// This sync operation will also complete them
		const callbacks = this._activeAsyncOperations.get(strPath);
		this._activeAsyncOperations.delete(strPath);

		// Run the operation
		// When in idle mode, we will enter sync mode
		let result;
		try {
			result = /** @type {Function***REMOVED*** */ (this._syncProvider).call(
				this._providerContext,
				path
			);
		***REMOVED*** catch (err) {
			this._storeResult(strPath, /** @type {Error***REMOVED*** */ (err), undefined);
			this._enterSyncModeWhenIdle();
			if (callbacks) {
				runCallbacks(callbacks, /** @type {Error***REMOVED*** */ (err), undefined);
			***REMOVED***
			throw err;
		***REMOVED***
		this._storeResult(strPath, null, result);
		this._enterSyncModeWhenIdle();
		if (callbacks) {
			runCallbacks(callbacks, null, result);
		***REMOVED***
		return result;
	***REMOVED***

	/**
	 * @param {string | Buffer | URL | number | (string | URL | Buffer | number)[] | Set<string | URL | Buffer | number>***REMOVED*** [what] what to purge
	 */
	purge(what) {
		if (!what) {
			if (this._mode !== STORAGE_MODE_IDLE) {
				this._data.clear();
				for (const level of this._levels) {
					level.clear();
				***REMOVED***
				this._enterIdleMode();
			***REMOVED***
		***REMOVED*** else if (
			typeof what === "string" ||
			Buffer.isBuffer(what) ||
			what instanceof URL ||
			typeof what === "number"
		) {
			const strWhat = typeof what !== "string" ? what.toString() : what;
			for (let [key, data] of this._data) {
				if (key.startsWith(strWhat)) {
					this._data.delete(key);
					data.level.delete(key);
				***REMOVED***
			***REMOVED***
			if (this._data.size === 0) {
				this._enterIdleMode();
			***REMOVED***
		***REMOVED*** else {
			for (let [key, data] of this._data) {
				for (const item of what) {
					const strItem = typeof item !== "string" ? item.toString() : item;
					if (key.startsWith(strItem)) {
						this._data.delete(key);
						data.level.delete(key);
						break;
					***REMOVED***
				***REMOVED***
			***REMOVED***
			if (this._data.size === 0) {
				this._enterIdleMode();
			***REMOVED***
		***REMOVED***
	***REMOVED***

	/**
	 * @param {string | Buffer | URL | number | (string | URL | Buffer | number)[] | Set<string | URL | Buffer | number>***REMOVED*** [what] what to purge
	 */
	purgeParent(what) {
		if (!what) {
			this.purge();
		***REMOVED*** else if (
			typeof what === "string" ||
			Buffer.isBuffer(what) ||
			what instanceof URL ||
			typeof what === "number"
		) {
			const strWhat = typeof what !== "string" ? what.toString() : what;
			this.purge(dirname(strWhat));
		***REMOVED*** else {
			const set = new Set();
			for (const item of what) {
				const strItem = typeof item !== "string" ? item.toString() : item;
				set.add(dirname(strItem));
			***REMOVED***
			this.purge(set);
		***REMOVED***
	***REMOVED***

	/**
	 * @param {string***REMOVED*** path path
	 * @param {Error | null***REMOVED*** err error
	 * @param {any***REMOVED*** result result
	 */
	_storeResult(path, err, result) {
		if (this._data.has(path)) return;
		const level = this._levels[this._currentLevel];
		this._data.set(path, { err, result, level ***REMOVED***);
		level.add(path);
	***REMOVED***

	_decayLevel() {
		const nextLevel = (this._currentLevel + 1) % this._levels.length;
		const decay = this._levels[nextLevel];
		this._currentLevel = nextLevel;
		for (let item of decay) {
			this._data.delete(item);
		***REMOVED***
		decay.clear();
		if (this._data.size === 0) {
			this._enterIdleMode();
		***REMOVED*** else {
			/** @type {number***REMOVED*** */
			(this._nextDecay) += this._tickInterval;
		***REMOVED***
	***REMOVED***

	_runDecays() {
		while (
			/** @type {number***REMOVED*** */ (this._nextDecay) <= Date.now() &&
			this._mode !== STORAGE_MODE_IDLE
		) {
			this._decayLevel();
		***REMOVED***
	***REMOVED***

	_enterAsyncMode() {
		let timeout = 0;
		switch (this._mode) {
			case STORAGE_MODE_ASYNC:
				return;
			case STORAGE_MODE_IDLE:
				this._nextDecay = Date.now() + this._tickInterval;
				timeout = this._tickInterval;
				break;
			case STORAGE_MODE_SYNC:
				this._runDecays();
				// _runDecays may change the mode
				if (
					/** @type {STORAGE_MODE_IDLE | STORAGE_MODE_SYNC | STORAGE_MODE_ASYNC***REMOVED****/
					(this._mode) === STORAGE_MODE_IDLE
				)
					return;
				timeout = Math.max(
					0,
					/** @type {number***REMOVED*** */ (this._nextDecay) - Date.now()
				);
				break;
		***REMOVED***
		this._mode = STORAGE_MODE_ASYNC;
		const ref = setTimeout(() => {
			this._mode = STORAGE_MODE_SYNC;
			this._runDecays();
		***REMOVED***, timeout);
		if (ref.unref) ref.unref();
		this._timeout = ref;
	***REMOVED***

	_enterSyncModeWhenIdle() {
		if (this._mode === STORAGE_MODE_IDLE) {
			this._mode = STORAGE_MODE_SYNC;
			this._nextDecay = Date.now() + this._tickInterval;
		***REMOVED***
	***REMOVED***

	_enterIdleMode() {
		this._mode = STORAGE_MODE_IDLE;
		this._nextDecay = undefined;
		if (this._timeout) clearTimeout(this._timeout);
	***REMOVED***
***REMOVED***

/**
 * @template {function***REMOVED*** Provider
 * @template {function***REMOVED*** AsyncProvider
 * @template FileSystem
 * @param {number***REMOVED*** duration duration in ms files are cached
 * @param {Provider | undefined***REMOVED*** provider provider
 * @param {AsyncProvider | undefined***REMOVED*** syncProvider sync provider
 * @param {BaseFileSystem***REMOVED*** providerContext provider context
 * @returns {OperationMergerBackend | CacheBackend***REMOVED*** backend
 */
const createBackend = (duration, provider, syncProvider, providerContext) => {
	if (duration > 0) {
		return new CacheBackend(duration, provider, syncProvider, providerContext);
	***REMOVED***
	return new OperationMergerBackend(provider, syncProvider, providerContext);
***REMOVED***;

module.exports = class CachedInputFileSystem {
	/**
	 * @param {BaseFileSystem***REMOVED*** fileSystem file system
	 * @param {number***REMOVED*** duration duration in ms files are cached
	 */
	constructor(fileSystem, duration) {
		this.fileSystem = fileSystem;

		this._lstatBackend = createBackend(
			duration,
			this.fileSystem.lstat,
			this.fileSystem.lstatSync,
			this.fileSystem
		);
		const lstat = this._lstatBackend.provide;
		this.lstat = /** @type {FileSystem["lstat"]***REMOVED*** */ (lstat);
		const lstatSync = this._lstatBackend.provideSync;
		this.lstatSync = /** @type {SyncFileSystem["lstatSync"]***REMOVED*** */ (lstatSync);

		this._statBackend = createBackend(
			duration,
			this.fileSystem.stat,
			this.fileSystem.statSync,
			this.fileSystem
		);
		const stat = this._statBackend.provide;
		this.stat = /** @type {FileSystem["stat"]***REMOVED*** */ (stat);
		const statSync = this._statBackend.provideSync;
		this.statSync = /** @type {SyncFileSystem["statSync"]***REMOVED*** */ (statSync);

		this._readdirBackend = createBackend(
			duration,
			this.fileSystem.readdir,
			this.fileSystem.readdirSync,
			this.fileSystem
		);
		const readdir = this._readdirBackend.provide;
		this.readdir = /** @type {FileSystem["readdir"]***REMOVED*** */ (readdir);
		const readdirSync = this._readdirBackend.provideSync;
		this.readdirSync = /** @type {SyncFileSystem["readdirSync"]***REMOVED*** */ (
			readdirSync
		);

		this._readFileBackend = createBackend(
			duration,
			this.fileSystem.readFile,
			this.fileSystem.readFileSync,
			this.fileSystem
		);
		const readFile = this._readFileBackend.provide;
		this.readFile = /** @type {FileSystem["readFile"]***REMOVED*** */ (readFile);
		const readFileSync = this._readFileBackend.provideSync;
		this.readFileSync = /** @type {SyncFileSystem["readFileSync"]***REMOVED*** */ (
			readFileSync
		);

		this._readJsonBackend = createBackend(
			duration,
			// prettier-ignore
			this.fileSystem.readJson ||
				(this.readFile &&
					(
						/**
						 * @param {string***REMOVED*** path path
						 * @param {FileSystemCallback<any>***REMOVED*** callback
						 */
						(path, callback) => {
							this.readFile(path, (err, buffer) => {
								if (err) return callback(err);
								if (!buffer || buffer.length === 0)
									return callback(new Error("No file content"));
								let data;
								try {
									data = JSON.parse(buffer.toString("utf-8"));
								***REMOVED*** catch (e) {
									return callback(/** @type {Error***REMOVED*** */ (e));
								***REMOVED***
								callback(null, data);
							***REMOVED***);
						***REMOVED***)
				),
			// prettier-ignore
			this.fileSystem.readJsonSync ||
				(this.readFileSync &&
					(
						/**
						 * @param {string***REMOVED*** path path
						 * @returns {any***REMOVED*** result
						 */
						(path) => {
							const buffer = this.readFileSync(path);
							const data = JSON.parse(buffer.toString("utf-8"));
							return data;
						***REMOVED***
				 )),
			this.fileSystem
		);
		const readJson = this._readJsonBackend.provide;
		this.readJson = /** @type {FileSystem["readJson"]***REMOVED*** */ (readJson);
		const readJsonSync = this._readJsonBackend.provideSync;
		this.readJsonSync = /** @type {SyncFileSystem["readJsonSync"]***REMOVED*** */ (
			readJsonSync
		);

		this._readlinkBackend = createBackend(
			duration,
			this.fileSystem.readlink,
			this.fileSystem.readlinkSync,
			this.fileSystem
		);
		const readlink = this._readlinkBackend.provide;
		this.readlink = /** @type {FileSystem["readlink"]***REMOVED*** */ (readlink);
		const readlinkSync = this._readlinkBackend.provideSync;
		this.readlinkSync = /** @type {SyncFileSystem["readlinkSync"]***REMOVED*** */ (
			readlinkSync
		);

		this._realpathBackend = createBackend(
			duration,
			this.fileSystem.realpath,
			this.fileSystem.realpathSync,
			this.fileSystem
		);
		const realpath = this._realpathBackend.provide;
		this.realpath = /** @type {FileSystem["realpath"]***REMOVED*** */ (realpath);
		const realpathSync = this._realpathBackend.provideSync;
		this.realpathSync = /** @type {SyncFileSystem["realpathSync"]***REMOVED*** */ (
			realpathSync
		);
	***REMOVED***

	/**
	 * @param {string | Buffer | URL | number | (string | URL | Buffer | number)[] | Set<string | URL | Buffer | number>***REMOVED*** [what] what to purge
	 */
	purge(what) {
		this._statBackend.purge(what);
		this._lstatBackend.purge(what);
		this._readdirBackend.purgeParent(what);
		this._readFileBackend.purge(what);
		this._readlinkBackend.purge(what);
		this._readJsonBackend.purge(what);
		this._realpathBackend.purge(what);
	***REMOVED***
***REMOVED***;
