// enhanced-resolve/lib/createInnerContext.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

/** @typedef {import("./Resolver").ResolveContext***REMOVED*** ResolveContext */

/**
 * @param {ResolveContext***REMOVED*** options options for inner context
 * @param {null|string***REMOVED*** message message to log
 * @returns {ResolveContext***REMOVED*** inner context
 */
module.exports = function createInnerContext(options, message) {
	let messageReported = false;
	let innerLog = undefined;
	if (options.log) {
		if (message) {
			/**
			 * @param {string***REMOVED*** msg message
			 */
			innerLog = msg => {
				if (!messageReported) {
					/** @type {(function(string): void)***REMOVED*** */
					(options.log)(message);
					messageReported = true;
				***REMOVED***

				/** @type {(function(string): void)***REMOVED*** */
				(options.log)("  " + msg);
			***REMOVED***;
		***REMOVED*** else {
			innerLog = options.log;
		***REMOVED***
	***REMOVED***

	return {
		log: innerLog,
		yield: options.yield,
		fileDependencies: options.fileDependencies,
		contextDependencies: options.contextDependencies,
		missingDependencies: options.missingDependencies,
		stack: options.stack
	***REMOVED***;
***REMOVED***;
