// enhanced-resolve/lib/ParsePlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

/** @typedef {import("./Resolver")***REMOVED*** Resolver */
/** @typedef {import("./Resolver").ResolveRequest***REMOVED*** ResolveRequest */
/** @typedef {import("./Resolver").ResolveStepHook***REMOVED*** ResolveStepHook */

module.exports = class ParsePlugin {
	/**
	 * @param {string | ResolveStepHook***REMOVED*** source source
	 * @param {Partial<ResolveRequest>***REMOVED*** requestOptions request options
	 * @param {string | ResolveStepHook***REMOVED*** target target
	 */
	constructor(source, requestOptions, target) {
		this.source = source;
		this.requestOptions = requestOptions;
		this.target = target;
	***REMOVED***

	/**
	 * @param {Resolver***REMOVED*** resolver the resolver
	 * @returns {void***REMOVED***
	 */
	apply(resolver) {
		const target = resolver.ensureHook(this.target);
		resolver
			.getHook(this.source)
			.tapAsync("ParsePlugin", (request, resolveContext, callback) => {
				const parsed = resolver.parse(/** @type {string***REMOVED*** */ (request.request));
				/** @type {ResolveRequest***REMOVED*** */
				const obj = { ...request, ...parsed, ...this.requestOptions ***REMOVED***;
				if (request.query && !parsed.query) {
					obj.query = request.query;
				***REMOVED***
				if (request.fragment && !parsed.fragment) {
					obj.fragment = request.fragment;
				***REMOVED***
				if (parsed && resolveContext.log) {
					if (parsed.module) resolveContext.log("Parsed request is a module");
					if (parsed.directory)
						resolveContext.log("Parsed request is a directory");
				***REMOVED***
				// There is an edge-case where a request with # can be a path or a fragment -> try both
				if (obj.request && !obj.query && obj.fragment) {
					const directory = obj.fragment.endsWith("/");
					/** @type {ResolveRequest***REMOVED*** */
					const alternative = {
						...obj,
						directory,
						request:
							obj.request +
							(obj.directory ? "/" : "") +
							(directory ? obj.fragment.slice(0, -1) : obj.fragment),
						fragment: ""
					***REMOVED***;
					resolver.doResolve(
						target,
						alternative,
						null,
						resolveContext,
						(err, result) => {
							if (err) return callback(err);
							if (result) return callback(null, result);
							resolver.doResolve(target, obj, null, resolveContext, callback);
						***REMOVED***
					);
					return;
				***REMOVED***
				resolver.doResolve(target, obj, null, resolveContext, callback);
			***REMOVED***);
	***REMOVED***
***REMOVED***;
