// enhanced-resolve/lib/PnpPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Maël Nison @arcanis
*/

"use strict";

/** @typedef {import("./Resolver")***REMOVED*** Resolver */
/** @typedef {import("./Resolver").ResolveStepHook***REMOVED*** ResolveStepHook */
/** @typedef {import("./Resolver").ResolveRequest***REMOVED*** ResolveRequest */
/**
 * @typedef {Object***REMOVED*** PnpApiImpl
 * @property {function(string, string, object): string | null***REMOVED*** resolveToUnqualified
 */

module.exports = class PnpPlugin {
	/**
	 * @param {string | ResolveStepHook***REMOVED*** source source
	 * @param {PnpApiImpl***REMOVED*** pnpApi pnpApi
	 * @param {string | ResolveStepHook***REMOVED*** target target
	 * @param {string | ResolveStepHook***REMOVED*** alternateTarget alternateTarget
	 */
	constructor(source, pnpApi, target, alternateTarget) {
		this.source = source;
		this.pnpApi = pnpApi;
		this.target = target;
		this.alternateTarget = alternateTarget;
	***REMOVED***

	/**
	 * @param {Resolver***REMOVED*** resolver the resolver
	 * @returns {void***REMOVED***
	 */
	apply(resolver) {
		/** @type {ResolveStepHook***REMOVED*** */
		const target = resolver.ensureHook(this.target);
		const alternateTarget = resolver.ensureHook(this.alternateTarget);
		resolver
			.getHook(this.source)
			.tapAsync("PnpPlugin", (request, resolveContext, callback) => {
				const req = request.request;
				if (!req) return callback();

				// The trailing slash indicates to PnP that this value is a folder rather than a file
				const issuer = `${request.path***REMOVED***/`;

				const packageMatch = /^(@[^/]+\/)?[^/]+/.exec(req);
				if (!packageMatch) return callback();

				const packageName = packageMatch[0];
				const innerRequest = `.${req.slice(packageName.length)***REMOVED***`;

				/** @type {string|undefined|null***REMOVED*** */
				let resolution;
				/** @type {string|undefined|null***REMOVED*** */
				let apiResolution;
				try {
					resolution = this.pnpApi.resolveToUnqualified(packageName, issuer, {
						considerBuiltins: false
					***REMOVED***);

					if (resolution === null) {
						// This is either not a PnP managed issuer or it's a Node builtin
						// Try to continue resolving with our alternatives
						resolver.doResolve(
							alternateTarget,
							request,
							"issuer is not managed by a pnpapi",
							resolveContext,
							(err, result) => {
								if (err) return callback(err);
								if (result) return callback(null, result);
								// Skip alternatives
								return callback(null, null);
							***REMOVED***
						);
						return;
					***REMOVED***

					if (resolveContext.fileDependencies) {
						apiResolution = this.pnpApi.resolveToUnqualified("pnpapi", issuer, {
							considerBuiltins: false
						***REMOVED***);
					***REMOVED***
				***REMOVED*** catch (/** @type {unknown***REMOVED*** */ error) {
					if (
						/** @type {Error & { code: string ***REMOVED******REMOVED*** */
						(error).code === "MODULE_NOT_FOUND" &&
						/** @type {Error & { pnpCode: string ***REMOVED******REMOVED*** */
						(error).pnpCode === "UNDECLARED_DEPENDENCY"
					) {
						// This is not a PnP managed dependency.
						// Try to continue resolving with our alternatives
						if (resolveContext.log) {
							resolveContext.log(`request is not managed by the pnpapi`);
							for (const line of /** @type {Error***REMOVED*** */ (error).message
								.split("\n")
								.filter(Boolean))
								resolveContext.log(`  ${line***REMOVED***`);
						***REMOVED***
						return callback();
					***REMOVED***
					return callback(/** @type {Error***REMOVED*** */ (error));
				***REMOVED***

				if (resolution === packageName) return callback();

				if (apiResolution && resolveContext.fileDependencies) {
					resolveContext.fileDependencies.add(apiResolution);
				***REMOVED***
				/** @type {ResolveRequest***REMOVED*** */
				const obj = {
					...request,
					path: resolution,
					request: innerRequest,
					ignoreSymlinks: true,
					fullySpecified: request.fullySpecified && innerRequest !== "."
				***REMOVED***;
				resolver.doResolve(
					target,
					obj,
					`resolved by pnp to ${resolution***REMOVED***`,
					resolveContext,
					(err, result) => {
						if (err) return callback(err);
						if (result) return callback(null, result);
						// Skip alternatives
						return callback(null, null);
					***REMOVED***
				);
			***REMOVED***);
	***REMOVED***
***REMOVED***;
