// enhanced-resolve/lib/RootsPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Ivan Kopeykin @vankop
*/

"use strict";

const forEachBail = require("./forEachBail");

/** @typedef {import("./Resolver")***REMOVED*** Resolver */
/** @typedef {import("./Resolver").ResolveRequest***REMOVED*** ResolveRequest */
/** @typedef {import("./Resolver").ResolveStepHook***REMOVED*** ResolveStepHook */

class RootsPlugin {
	/**
	 * @param {string | ResolveStepHook***REMOVED*** source source hook
	 * @param {Set<string>***REMOVED*** roots roots
	 * @param {string | ResolveStepHook***REMOVED*** target target hook
	 */
	constructor(source, roots, target) {
		this.roots = Array.from(roots);
		this.source = source;
		this.target = target;
	***REMOVED***

	/**
	 * @param {Resolver***REMOVED*** resolver the resolver
	 * @returns {void***REMOVED***
	 */
	apply(resolver) {
		const target = resolver.ensureHook(this.target);

		resolver
			.getHook(this.source)
			.tapAsync("RootsPlugin", (request, resolveContext, callback) => {
				const req = request.request;
				if (!req) return callback();
				if (!req.startsWith("/")) return callback();

				forEachBail(
					this.roots,
					/**
					 * @param {string***REMOVED*** root root
					 * @param {(err?: null|Error, result?: null|ResolveRequest) => void***REMOVED*** callback callback
					 * @returns {void***REMOVED***
					 */
					(root, callback) => {
						const path = resolver.join(root, req.slice(1));
						/** @type {ResolveRequest***REMOVED*** */
						const obj = {
							...request,
							path,
							relativePath: request.relativePath && path
						***REMOVED***;
						resolver.doResolve(
							target,
							obj,
							`root path ${root***REMOVED***`,
							resolveContext,
							callback
						);
					***REMOVED***,
					callback
				);
			***REMOVED***);
	***REMOVED***
***REMOVED***

module.exports = RootsPlugin;
