// enhanced-resolve/lib/SelfReferencePlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const DescriptionFileUtils = require("./DescriptionFileUtils");

/** @typedef {import("./Resolver")***REMOVED*** Resolver */
/** @typedef {import("./Resolver").JsonObject***REMOVED*** JsonObject */
/** @typedef {import("./Resolver").ResolveRequest***REMOVED*** ResolveRequest */
/** @typedef {import("./Resolver").ResolveStepHook***REMOVED*** ResolveStepHook */

const slashCode = "/".charCodeAt(0);

module.exports = class SelfReferencePlugin {
	/**
	 * @param {string | ResolveStepHook***REMOVED*** source source
	 * @param {string | string[]***REMOVED*** fieldNamePath name path
	 * @param {string | ResolveStepHook***REMOVED*** target target
	 */
	constructor(source, fieldNamePath, target) {
		this.source = source;
		this.target = target;
		this.fieldName = fieldNamePath;
	***REMOVED***

	/**
	 * @param {Resolver***REMOVED*** resolver the resolver
	 * @returns {void***REMOVED***
	 */
	apply(resolver) {
		const target = resolver.ensureHook(this.target);
		resolver
			.getHook(this.source)
			.tapAsync("SelfReferencePlugin", (request, resolveContext, callback) => {
				if (!request.descriptionFilePath) return callback();

				const req = request.request;
				if (!req) return callback();

				// Feature is only enabled when an exports field is present
				const exportsField = DescriptionFileUtils.getField(
					/** @type {JsonObject***REMOVED*** */ (request.descriptionFileData),
					this.fieldName
				);
				if (!exportsField) return callback();

				const name = DescriptionFileUtils.getField(
					/** @type {JsonObject***REMOVED*** */ (request.descriptionFileData),
					"name"
				);
				if (typeof name !== "string") return callback();

				if (
					req.startsWith(name) &&
					(req.length === name.length ||
						req.charCodeAt(name.length) === slashCode)
				) {
					const remainingRequest = `.${req.slice(name.length)***REMOVED***`;
					/** @type {ResolveRequest***REMOVED*** */
					const obj = {
						...request,
						request: remainingRequest,
						path: /** @type {string***REMOVED*** */ (request.descriptionFileRoot),
						relativePath: "."
					***REMOVED***;

					resolver.doResolve(
						target,
						obj,
						"self reference",
						resolveContext,
						callback
					);
				***REMOVED*** else {
					return callback();
				***REMOVED***
			***REMOVED***);
	***REMOVED***
***REMOVED***;
