// enhanced-resolve/lib/SyncAsyncFileSystemDecorator.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

/** @typedef {import("./Resolver").FileSystem***REMOVED*** FileSystem */
/** @typedef {import("./Resolver").ReaddirStringCallback***REMOVED*** ReaddirStringCallback */
/** @typedef {import("./Resolver").StringCallback***REMOVED*** StringCallback */
/** @typedef {import("./Resolver").SyncFileSystem***REMOVED*** SyncFileSystem */

/**
 * @param {SyncFileSystem***REMOVED*** fs file system implementation
 * @constructor
 */
function SyncAsyncFileSystemDecorator(fs) {
	this.fs = fs;

	this.lstat = undefined;
	this.lstatSync = undefined;
	const lstatSync = fs.lstatSync;
	if (lstatSync) {
		this.lstat =
			/** @type {FileSystem["lstat"]***REMOVED*** */
			(
				(arg, options, callback) => {
					let result;
					try {
						result = /** @type {Function | undefined***REMOVED*** */ (callback)
							? lstatSync.call(fs, arg, options)
							: lstatSync.call(fs, arg);
					***REMOVED*** catch (e) {
						return (callback || options)(
							/** @type {NodeJS.ErrnoException | null***REMOVED*** */ (e)
						);
					***REMOVED***

					(callback || options)(null, /** @type {any***REMOVED*** */ (result));
				***REMOVED***
			);
		this.lstatSync =
			/** @type {SyncFileSystem["lstatSync"]***REMOVED*** */
			((arg, options) => lstatSync.call(fs, arg, options));
	***REMOVED***

	this.stat =
		/** @type {FileSystem["stat"]***REMOVED*** */
		(
			(arg, options, callback) => {
				let result;
				try {
					result = /** @type {Function | undefined***REMOVED*** */ (callback)
						? fs.statSync(arg, options)
						: fs.statSync(arg);
				***REMOVED*** catch (e) {
					return (callback || options)(
						/** @type {NodeJS.ErrnoException | null***REMOVED*** */ (e)
					);
				***REMOVED***

				(callback || options)(null, /** @type {any***REMOVED*** */ (result));
			***REMOVED***
		);
	this.statSync =
		/** @type {SyncFileSystem["statSync"]***REMOVED*** */
		((arg, options) => fs.statSync(arg, options));

	this.readdir =
		/** @type {FileSystem["readdir"]***REMOVED*** */
		(
			(arg, options, callback) => {
				let result;
				try {
					result = /** @type {Function | undefined***REMOVED*** */ (callback)
						? fs.readdirSync(
								arg,
								/** @type {Exclude<Parameters<FileSystem["readdir"]>[1], ReaddirStringCallback>***REMOVED*** */
								(options)
						  )
						: fs.readdirSync(arg);
				***REMOVED*** catch (e) {
					return (callback || options)(
						/** @type {NodeJS.ErrnoException | null***REMOVED*** */ (e)
					);
				***REMOVED***

				(callback || options)(null, /** @type {any***REMOVED*** */ (result));
			***REMOVED***
		);
	this.readdirSync =
		/** @type {SyncFileSystem["readdirSync"]***REMOVED*** */
		(
			(arg, options) =>
				fs.readdirSync(
					arg,
					/** @type {Parameters<SyncFileSystem["readdirSync"]>[1]***REMOVED*** */ (options)
				)
		);

	this.readFile =
		/** @type {FileSystem["readFile"]***REMOVED*** */
		(
			(arg, options, callback) => {
				let result;
				try {
					result = /** @type {Function | undefined***REMOVED*** */ (callback)
						? fs.readFileSync(arg, options)
						: fs.readFileSync(arg);
				***REMOVED*** catch (e) {
					return (callback || options)(
						/** @type {NodeJS.ErrnoException | null***REMOVED*** */ (e)
					);
				***REMOVED***

				(callback || options)(null, /** @type {any***REMOVED*** */ (result));
			***REMOVED***
		);
	this.readFileSync =
		/** @type {SyncFileSystem["readFileSync"]***REMOVED*** */
		((arg, options) => fs.readFileSync(arg, options));

	this.readlink =
		/** @type {FileSystem["readlink"]***REMOVED*** */
		(
			(arg, options, callback) => {
				let result;
				try {
					result = /** @type {Function | undefined***REMOVED*** */ (callback)
						? fs.readlinkSync(
								arg,
								/** @type {Exclude<Parameters<FileSystem["readlink"]>[1], StringCallback>***REMOVED*** */
								(options)
						  )
						: fs.readlinkSync(arg);
				***REMOVED*** catch (e) {
					return (callback || options)(
						/** @type {NodeJS.ErrnoException | null***REMOVED*** */ (e)
					);
				***REMOVED***

				(callback || options)(null, /** @type {any***REMOVED*** */ (result));
			***REMOVED***
		);
	this.readlinkSync =
		/** @type {SyncFileSystem["readlinkSync"]***REMOVED*** */
		(
			(arg, options) =>
				fs.readlinkSync(
					arg,
					/** @type {Parameters<SyncFileSystem["readlinkSync"]>[1]***REMOVED*** */ (options)
				)
		);

	this.readJson = undefined;
	this.readJsonSync = undefined;
	const readJsonSync = fs.readJsonSync;
	if (readJsonSync) {
		this.readJson =
			/** @type {FileSystem["readJson"]***REMOVED*** */
			(
				(arg, callback) => {
					let result;
					try {
						result = readJsonSync.call(fs, arg);
					***REMOVED*** catch (e) {
						return callback(
							/** @type {NodeJS.ErrnoException | Error | null***REMOVED*** */ (e)
						);
					***REMOVED***

					callback(null, result);
				***REMOVED***
			);
		this.readJsonSync =
			/** @type {SyncFileSystem["readJsonSync"]***REMOVED*** */
			(arg => readJsonSync.call(fs, arg));
	***REMOVED***

	this.realpath = undefined;
	this.realpathSync = undefined;
	const realpathSync = fs.realpathSync;
	if (realpathSync) {
		this.realpath =
			/** @type {FileSystem["realpath"]***REMOVED*** */
			(
				(arg, options, callback) => {
					let result;
					try {
						result = /** @type {Function | undefined***REMOVED*** */ (callback)
							? realpathSync.call(
									fs,
									arg,
									/** @type {Exclude<Parameters<NonNullable<FileSystem["realpath"]>>[1], StringCallback>***REMOVED*** */
									(options)
							  )
							: realpathSync.call(fs, arg);
					***REMOVED*** catch (e) {
						return (callback || options)(
							/** @type {NodeJS.ErrnoException | null***REMOVED*** */ (e)
						);
					***REMOVED***

					(callback || options)(null, /** @type {any***REMOVED*** */ (result));
				***REMOVED***
			);
		this.realpathSync =
			/** @type {SyncFileSystem["realpathSync"]***REMOVED*** */
			(
				(arg, options) =>
					realpathSync.call(
						fs,
						arg,
						/** @type {Parameters<NonNullable<SyncFileSystem["realpathSync"]>>[1]***REMOVED*** */
						(options)
					)
			);
	***REMOVED***
***REMOVED***
module.exports = SyncAsyncFileSystemDecorator;
