// json5/lib/parse.js
const util = require('./util')

let source
let parseState
let stack
let pos
let line
let column
let token
let key
let root

module.exports = function parse (text, reviver) {
    source = String(text)
    parseState = 'start'
    stack = []
    pos = 0
    line = 1
    column = 0
    token = undefined
    key = undefined
    root = undefined

    do {
        token = lex()

        // This code is unreachable.
        // if (!parseStates[parseState]) {
        //     throw invalidParseState()
        // ***REMOVED***

        parseStates[parseState]()
    ***REMOVED*** while (token.type !== 'eof')

    if (typeof reviver === 'function') {
        return internalize({'': root***REMOVED***, '', reviver)
    ***REMOVED***

    return root
***REMOVED***

function internalize (holder, name, reviver) {
    const value = holder[name]
    if (value != null && typeof value === 'object') {
        if (Array.isArray(value)) {
            for (let i = 0; i < value.length; i++) {
                const key = String(i)
                const replacement = internalize(value, key, reviver)
                if (replacement === undefined) {
                    delete value[key]
                ***REMOVED*** else {
                    Object.defineProperty(value, key, {
                        value: replacement,
                        writable: true,
                        enumerable: true,
                        configurable: true,
                    ***REMOVED***)
                ***REMOVED***
            ***REMOVED***
        ***REMOVED*** else {
            for (const key in value) {
                const replacement = internalize(value, key, reviver)
                if (replacement === undefined) {
                    delete value[key]
                ***REMOVED*** else {
                    Object.defineProperty(value, key, {
                        value: replacement,
                        writable: true,
                        enumerable: true,
                        configurable: true,
                    ***REMOVED***)
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    return reviver.call(holder, name, value)
***REMOVED***

let lexState
let buffer
let doubleQuote
let sign
let c

function lex () {
    lexState = 'default'
    buffer = ''
    doubleQuote = false
    sign = 1

    for (;;) {
        c = peek()

        // This code is unreachable.
        // if (!lexStates[lexState]) {
        //     throw invalidLexState(lexState)
        // ***REMOVED***

        const token = lexStates[lexState]()
        if (token) {
            return token
        ***REMOVED***
    ***REMOVED***
***REMOVED***

function peek () {
    if (source[pos]) {
        return String.fromCodePoint(source.codePointAt(pos))
    ***REMOVED***
***REMOVED***

function read () {
    const c = peek()

    if (c === '\n') {
        line++
        column = 0
    ***REMOVED*** else if (c) {
        column += c.length
    ***REMOVED*** else {
        column++
    ***REMOVED***

    if (c) {
        pos += c.length
    ***REMOVED***

    return c
***REMOVED***

const lexStates = {
    default () {
        switch (c) {
        case '\t':
        case '\v':
        case '\f':
        case ' ':
        case '\u00A0':
        case '\uFEFF':
        case '\n':
        case '\r':
        case '\u2028':
        case '\u2029':
            read()
            return

        case '/':
            read()
            lexState = 'comment'
            return

        case undefined:
            read()
            return newToken('eof')
        ***REMOVED***

        if (util.isSpaceSeparator(c)) {
            read()
            return
        ***REMOVED***

        // This code is unreachable.
        // if (!lexStates[parseState]) {
        //     throw invalidLexState(parseState)
        // ***REMOVED***

        return lexStates[parseState]()
    ***REMOVED***,

    comment () {
        switch (c) {
        case '*':
            read()
            lexState = 'multiLineComment'
            return

        case '/':
            read()
            lexState = 'singleLineComment'
            return
        ***REMOVED***

        throw invalidChar(read())
    ***REMOVED***,

    multiLineComment () {
        switch (c) {
        case '*':
            read()
            lexState = 'multiLineCommentAsterisk'
            return

        case undefined:
            throw invalidChar(read())
        ***REMOVED***

        read()
    ***REMOVED***,

    multiLineCommentAsterisk () {
        switch (c) {
        case '*':
            read()
            return

        case '/':
            read()
            lexState = 'default'
            return

        case undefined:
            throw invalidChar(read())
        ***REMOVED***

        read()
        lexState = 'multiLineComment'
    ***REMOVED***,

    singleLineComment () {
        switch (c) {
        case '\n':
        case '\r':
        case '\u2028':
        case '\u2029':
            read()
            lexState = 'default'
            return

        case undefined:
            read()
            return newToken('eof')
        ***REMOVED***

        read()
    ***REMOVED***,

    value () {
        switch (c) {
        case '{':
        case '[':
            return newToken('punctuator', read())

        case 'n':
            read()
            literal('ull')
            return newToken('null', null)

        case 't':
            read()
            literal('rue')
            return newToken('boolean', true)

        case 'f':
            read()
            literal('alse')
            return newToken('boolean', false)

        case '-':
        case '+':
            if (read() === '-') {
                sign = -1
            ***REMOVED***

            lexState = 'sign'
            return

        case '.':
            buffer = read()
            lexState = 'decimalPointLeading'
            return

        case '0':
            buffer = read()
            lexState = 'zero'
            return

        case '1':
        case '2':
        case '3':
        case '4':
        case '5':
        case '6':
        case '7':
        case '8':
        case '9':
            buffer = read()
            lexState = 'decimalInteger'
            return

        case 'I':
            read()
            literal('nfinity')
            return newToken('numeric', Infinity)

        case 'N':
            read()
            literal('aN')
            return newToken('numeric', NaN)

        case '"':
        case "'":
            doubleQuote = (read() === '"')
            buffer = ''
            lexState = 'string'
            return
        ***REMOVED***

        throw invalidChar(read())
    ***REMOVED***,

    identifierNameStartEscape () {
        if (c !== 'u') {
            throw invalidChar(read())
        ***REMOVED***

        read()
        const u = unicodeEscape()
        switch (u) {
        case '$':
        case '_':
            break

        default:
            if (!util.isIdStartChar(u)) {
                throw invalidIdentifier()
            ***REMOVED***

            break
        ***REMOVED***

        buffer += u
        lexState = 'identifierName'
    ***REMOVED***,

    identifierName () {
        switch (c) {
        case '$':
        case '_':
        case '\u200C':
        case '\u200D':
            buffer += read()
            return

        case '\\':
            read()
            lexState = 'identifierNameEscape'
            return
        ***REMOVED***

        if (util.isIdContinueChar(c)) {
            buffer += read()
            return
        ***REMOVED***

        return newToken('identifier', buffer)
    ***REMOVED***,

    identifierNameEscape () {
        if (c !== 'u') {
            throw invalidChar(read())
        ***REMOVED***

        read()
        const u = unicodeEscape()
        switch (u) {
        case '$':
        case '_':
        case '\u200C':
        case '\u200D':
            break

        default:
            if (!util.isIdContinueChar(u)) {
                throw invalidIdentifier()
            ***REMOVED***

            break
        ***REMOVED***

        buffer += u
        lexState = 'identifierName'
    ***REMOVED***,

    sign () {
        switch (c) {
        case '.':
            buffer = read()
            lexState = 'decimalPointLeading'
            return

        case '0':
            buffer = read()
            lexState = 'zero'
            return

        case '1':
        case '2':
        case '3':
        case '4':
        case '5':
        case '6':
        case '7':
        case '8':
        case '9':
            buffer = read()
            lexState = 'decimalInteger'
            return

        case 'I':
            read()
            literal('nfinity')
            return newToken('numeric', sign * Infinity)

        case 'N':
            read()
            literal('aN')
            return newToken('numeric', NaN)
        ***REMOVED***

        throw invalidChar(read())
    ***REMOVED***,

    zero () {
        switch (c) {
        case '.':
            buffer += read()
            lexState = 'decimalPoint'
            return

        case 'e':
        case 'E':
            buffer += read()
            lexState = 'decimalExponent'
            return

        case 'x':
        case 'X':
            buffer += read()
            lexState = 'hexadecimal'
            return
        ***REMOVED***

        return newToken('numeric', sign * 0)
    ***REMOVED***,

    decimalInteger () {
        switch (c) {
        case '.':
            buffer += read()
            lexState = 'decimalPoint'
            return

        case 'e':
        case 'E':
            buffer += read()
            lexState = 'decimalExponent'
            return
        ***REMOVED***

        if (util.isDigit(c)) {
            buffer += read()
            return
        ***REMOVED***

        return newToken('numeric', sign * Number(buffer))
    ***REMOVED***,

    decimalPointLeading () {
        if (util.isDigit(c)) {
            buffer += read()
            lexState = 'decimalFraction'
            return
        ***REMOVED***

        throw invalidChar(read())
    ***REMOVED***,

    decimalPoint () {
        switch (c) {
        case 'e':
        case 'E':
            buffer += read()
            lexState = 'decimalExponent'
            return
        ***REMOVED***

        if (util.isDigit(c)) {
            buffer += read()
            lexState = 'decimalFraction'
            return
        ***REMOVED***

        return newToken('numeric', sign * Number(buffer))
    ***REMOVED***,

    decimalFraction () {
        switch (c) {
        case 'e':
        case 'E':
            buffer += read()
            lexState = 'decimalExponent'
            return
        ***REMOVED***

        if (util.isDigit(c)) {
            buffer += read()
            return
        ***REMOVED***

        return newToken('numeric', sign * Number(buffer))
    ***REMOVED***,

    decimalExponent () {
        switch (c) {
        case '+':
        case '-':
            buffer += read()
            lexState = 'decimalExponentSign'
            return
        ***REMOVED***

        if (util.isDigit(c)) {
            buffer += read()
            lexState = 'decimalExponentInteger'
            return
        ***REMOVED***

        throw invalidChar(read())
    ***REMOVED***,

    decimalExponentSign () {
        if (util.isDigit(c)) {
            buffer += read()
            lexState = 'decimalExponentInteger'
            return
        ***REMOVED***

        throw invalidChar(read())
    ***REMOVED***,

    decimalExponentInteger () {
        if (util.isDigit(c)) {
            buffer += read()
            return
        ***REMOVED***

        return newToken('numeric', sign * Number(buffer))
    ***REMOVED***,

    hexadecimal () {
        if (util.isHexDigit(c)) {
            buffer += read()
            lexState = 'hexadecimalInteger'
            return
        ***REMOVED***

        throw invalidChar(read())
    ***REMOVED***,

    hexadecimalInteger () {
        if (util.isHexDigit(c)) {
            buffer += read()
            return
        ***REMOVED***

        return newToken('numeric', sign * Number(buffer))
    ***REMOVED***,

    string () {
        switch (c) {
        case '\\':
            read()
            buffer += escape()
            return

        case '"':
            if (doubleQuote) {
                read()
                return newToken('string', buffer)
            ***REMOVED***

            buffer += read()
            return

        case "'":
            if (!doubleQuote) {
                read()
                return newToken('string', buffer)
            ***REMOVED***

            buffer += read()
            return

        case '\n':
        case '\r':
            throw invalidChar(read())

        case '\u2028':
        case '\u2029':
            separatorChar(c)
            break

        case undefined:
            throw invalidChar(read())
        ***REMOVED***

        buffer += read()
    ***REMOVED***,

    start () {
        switch (c) {
        case '{':
        case '[':
            return newToken('punctuator', read())

        // This code is unreachable since the default lexState handles eof.
        // case undefined:
        //     return newToken('eof')
        ***REMOVED***

        lexState = 'value'
    ***REMOVED***,

    beforePropertyName () {
        switch (c) {
        case '$':
        case '_':
            buffer = read()
            lexState = 'identifierName'
            return

        case '\\':
            read()
            lexState = 'identifierNameStartEscape'
            return

        case '***REMOVED***':
            return newToken('punctuator', read())

        case '"':
        case "'":
            doubleQuote = (read() === '"')
            lexState = 'string'
            return
        ***REMOVED***

        if (util.isIdStartChar(c)) {
            buffer += read()
            lexState = 'identifierName'
            return
        ***REMOVED***

        throw invalidChar(read())
    ***REMOVED***,

    afterPropertyName () {
        if (c === ':') {
            return newToken('punctuator', read())
        ***REMOVED***

        throw invalidChar(read())
    ***REMOVED***,

    beforePropertyValue () {
        lexState = 'value'
    ***REMOVED***,

    afterPropertyValue () {
        switch (c) {
        case ',':
        case '***REMOVED***':
            return newToken('punctuator', read())
        ***REMOVED***

        throw invalidChar(read())
    ***REMOVED***,

    beforeArrayValue () {
        if (c === ']') {
            return newToken('punctuator', read())
        ***REMOVED***

        lexState = 'value'
    ***REMOVED***,

    afterArrayValue () {
        switch (c) {
        case ',':
        case ']':
            return newToken('punctuator', read())
        ***REMOVED***

        throw invalidChar(read())
    ***REMOVED***,

    end () {
        // This code is unreachable since it's handled by the default lexState.
        // if (c === undefined) {
        //     read()
        //     return newToken('eof')
        // ***REMOVED***

        throw invalidChar(read())
    ***REMOVED***,
***REMOVED***

function newToken (type, value) {
    return {
        type,
        value,
        line,
        column,
    ***REMOVED***
***REMOVED***

function literal (s) {
    for (const c of s) {
        const p = peek()

        if (p !== c) {
            throw invalidChar(read())
        ***REMOVED***

        read()
    ***REMOVED***
***REMOVED***

function escape () {
    const c = peek()
    switch (c) {
    case 'b':
        read()
        return '\b'

    case 'f':
        read()
        return '\f'

    case 'n':
        read()
        return '\n'

    case 'r':
        read()
        return '\r'

    case 't':
        read()
        return '\t'

    case 'v':
        read()
        return '\v'

    case '0':
        read()
        if (util.isDigit(peek())) {
            throw invalidChar(read())
        ***REMOVED***

        return '\0'

    case 'x':
        read()
        return hexEscape()

    case 'u':
        read()
        return unicodeEscape()

    case '\n':
    case '\u2028':
    case '\u2029':
        read()
        return ''

    case '\r':
        read()
        if (peek() === '\n') {
            read()
        ***REMOVED***

        return ''

    case '1':
    case '2':
    case '3':
    case '4':
    case '5':
    case '6':
    case '7':
    case '8':
    case '9':
        throw invalidChar(read())

    case undefined:
        throw invalidChar(read())
    ***REMOVED***

    return read()
***REMOVED***

function hexEscape () {
    let buffer = ''
    let c = peek()

    if (!util.isHexDigit(c)) {
        throw invalidChar(read())
    ***REMOVED***

    buffer += read()

    c = peek()
    if (!util.isHexDigit(c)) {
        throw invalidChar(read())
    ***REMOVED***

    buffer += read()

    return String.fromCodePoint(parseInt(buffer, 16))
***REMOVED***

function unicodeEscape () {
    let buffer = ''
    let count = 4

    while (count-- > 0) {
        const c = peek()
        if (!util.isHexDigit(c)) {
            throw invalidChar(read())
        ***REMOVED***

        buffer += read()
    ***REMOVED***

    return String.fromCodePoint(parseInt(buffer, 16))
***REMOVED***

const parseStates = {
    start () {
        if (token.type === 'eof') {
            throw invalidEOF()
        ***REMOVED***

        push()
    ***REMOVED***,

    beforePropertyName () {
        switch (token.type) {
        case 'identifier':
        case 'string':
            key = token.value
            parseState = 'afterPropertyName'
            return

        case 'punctuator':
            // This code is unreachable since it's handled by the lexState.
            // if (token.value !== '***REMOVED***') {
            //     throw invalidToken()
            // ***REMOVED***

            pop()
            return

        case 'eof':
            throw invalidEOF()
        ***REMOVED***

        // This code is unreachable since it's handled by the lexState.
        // throw invalidToken()
    ***REMOVED***,

    afterPropertyName () {
        // This code is unreachable since it's handled by the lexState.
        // if (token.type !== 'punctuator' || token.value !== ':') {
        //     throw invalidToken()
        // ***REMOVED***

        if (token.type === 'eof') {
            throw invalidEOF()
        ***REMOVED***

        parseState = 'beforePropertyValue'
    ***REMOVED***,

    beforePropertyValue () {
        if (token.type === 'eof') {
            throw invalidEOF()
        ***REMOVED***

        push()
    ***REMOVED***,

    beforeArrayValue () {
        if (token.type === 'eof') {
            throw invalidEOF()
        ***REMOVED***

        if (token.type === 'punctuator' && token.value === ']') {
            pop()
            return
        ***REMOVED***

        push()
    ***REMOVED***,

    afterPropertyValue () {
        // This code is unreachable since it's handled by the lexState.
        // if (token.type !== 'punctuator') {
        //     throw invalidToken()
        // ***REMOVED***

        if (token.type === 'eof') {
            throw invalidEOF()
        ***REMOVED***

        switch (token.value) {
        case ',':
            parseState = 'beforePropertyName'
            return

        case '***REMOVED***':
            pop()
        ***REMOVED***

        // This code is unreachable since it's handled by the lexState.
        // throw invalidToken()
    ***REMOVED***,

    afterArrayValue () {
        // This code is unreachable since it's handled by the lexState.
        // if (token.type !== 'punctuator') {
        //     throw invalidToken()
        // ***REMOVED***

        if (token.type === 'eof') {
            throw invalidEOF()
        ***REMOVED***

        switch (token.value) {
        case ',':
            parseState = 'beforeArrayValue'
            return

        case ']':
            pop()
        ***REMOVED***

        // This code is unreachable since it's handled by the lexState.
        // throw invalidToken()
    ***REMOVED***,

    end () {
        // This code is unreachable since it's handled by the lexState.
        // if (token.type !== 'eof') {
        //     throw invalidToken()
        // ***REMOVED***
    ***REMOVED***,
***REMOVED***

function push () {
    let value

    switch (token.type) {
    case 'punctuator':
        switch (token.value) {
        case '{':
            value = {***REMOVED***
            break

        case '[':
            value = []
            break
        ***REMOVED***

        break

    case 'null':
    case 'boolean':
    case 'numeric':
    case 'string':
        value = token.value
        break

    // This code is unreachable.
    // default:
    //     throw invalidToken()
    ***REMOVED***

    if (root === undefined) {
        root = value
    ***REMOVED*** else {
        const parent = stack[stack.length - 1]
        if (Array.isArray(parent)) {
            parent.push(value)
        ***REMOVED*** else {
            Object.defineProperty(parent, key, {
                value,
                writable: true,
                enumerable: true,
                configurable: true,
            ***REMOVED***)
        ***REMOVED***
    ***REMOVED***

    if (value !== null && typeof value === 'object') {
        stack.push(value)

        if (Array.isArray(value)) {
            parseState = 'beforeArrayValue'
        ***REMOVED*** else {
            parseState = 'beforePropertyName'
        ***REMOVED***
    ***REMOVED*** else {
        const current = stack[stack.length - 1]
        if (current == null) {
            parseState = 'end'
        ***REMOVED*** else if (Array.isArray(current)) {
            parseState = 'afterArrayValue'
        ***REMOVED*** else {
            parseState = 'afterPropertyValue'
        ***REMOVED***
    ***REMOVED***
***REMOVED***

function pop () {
    stack.pop()

    const current = stack[stack.length - 1]
    if (current == null) {
        parseState = 'end'
    ***REMOVED*** else if (Array.isArray(current)) {
        parseState = 'afterArrayValue'
    ***REMOVED*** else {
        parseState = 'afterPropertyValue'
    ***REMOVED***
***REMOVED***

// This code is unreachable.
// function invalidParseState () {
//     return new Error(`JSON5: invalid parse state '${parseState***REMOVED***'`)
// ***REMOVED***

// This code is unreachable.
// function invalidLexState (state) {
//     return new Error(`JSON5: invalid lex state '${state***REMOVED***'`)
// ***REMOVED***

function invalidChar (c) {
    if (c === undefined) {
        return syntaxError(`JSON5: invalid end of input at ${line***REMOVED***:${column***REMOVED***`)
    ***REMOVED***

    return syntaxError(`JSON5: invalid character '${formatChar(c)***REMOVED***' at ${line***REMOVED***:${column***REMOVED***`)
***REMOVED***

function invalidEOF () {
    return syntaxError(`JSON5: invalid end of input at ${line***REMOVED***:${column***REMOVED***`)
***REMOVED***

// This code is unreachable.
// function invalidToken () {
//     if (token.type === 'eof') {
//         return syntaxError(`JSON5: invalid end of input at ${line***REMOVED***:${column***REMOVED***`)
//     ***REMOVED***

//     const c = String.fromCodePoint(token.value.codePointAt(0))
//     return syntaxError(`JSON5: invalid character '${formatChar(c)***REMOVED***' at ${line***REMOVED***:${column***REMOVED***`)
// ***REMOVED***

function invalidIdentifier () {
    column -= 5
    return syntaxError(`JSON5: invalid identifier character at ${line***REMOVED***:${column***REMOVED***`)
***REMOVED***

function separatorChar (c) {
    console.warn(`JSON5: '${formatChar(c)***REMOVED***' in strings is not valid ECMAScript; consider escaping`)
***REMOVED***

function formatChar (c) {
    const replacements = {
        "'": "\\'",
        '"': '\\"',
        '\\': '\\\\',
        '\b': '\\b',
        '\f': '\\f',
        '\n': '\\n',
        '\r': '\\r',
        '\t': '\\t',
        '\v': '\\v',
        '\0': '\\0',
        '\u2028': '\\u2028',
        '\u2029': '\\u2029',
    ***REMOVED***

    if (replacements[c]) {
        return replacements[c]
    ***REMOVED***

    if (c < ' ') {
        const hexString = c.charCodeAt(0).toString(16)
        return '\\x' + ('00' + hexString).substring(hexString.length)
    ***REMOVED***

    return c
***REMOVED***

function syntaxError (message) {
    const err = new SyntaxError(message)
    err.lineNumber = line
    err.columnNumber = column
    return err
***REMOVED***
