// @svgr/webpack/lib/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _loaderUtils = require("loader-utils");

var _core = require("@babel/core");

var _core2 = _interopRequireDefault(require("@svgr/core"));

var _pluginSvgo = _interopRequireDefault(require("@svgr/plugin-svgo"));

var _pluginJsx = _interopRequireDefault(require("@svgr/plugin-jsx"));

var _presetReact = _interopRequireDefault(require("@babel/preset-react"));

var _presetEnv = _interopRequireDefault(require("@babel/preset-env"));

var _pluginTransformReactConstantElements = _interopRequireDefault(require("@babel/plugin-transform-react-constant-elements"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj ***REMOVED***; ***REMOVED***

const babelOptions = {
  babelrc: false,
  configFile: false,
  presets: [(0, _core.createConfigItem)(_presetReact.default, {
    type: 'preset'
  ***REMOVED***), (0, _core.createConfigItem)([_presetEnv.default, {
    modules: false
  ***REMOVED***], {
    type: 'preset'
  ***REMOVED***)],
  plugins: [(0, _core.createConfigItem)(_pluginTransformReactConstantElements.default)]
***REMOVED***;

function svgrLoader(source) {
  const callback = this.async();
  const {
    babel = true,
    ...options
  ***REMOVED*** = (0, _loaderUtils.getOptions)(this) || {***REMOVED***;

  const readSvg = () => new Promise((resolve, reject) => {
    this.fs.readFile(this.resourcePath, (err, result) => {
      if (err) reject(err);
      resolve(result);
    ***REMOVED***);
  ***REMOVED***);

  const previousExport = (() => {
    if (source.toString('utf-8').startsWith('export ')) {
      return source;
    ***REMOVED***

    const exportMatches = source.toString('utf-8').match(/^module.exports\s*=\s*(.*)/);
    return exportMatches ? `export default ${exportMatches[1]***REMOVED***` : null;
  ***REMOVED***)();

  const tranformSvg = svg => (0, _core2.default)(svg, options, {
    caller: {
      name: '@svgr/webpack',
      previousExport,
      defaultPlugins: [_pluginSvgo.default, _pluginJsx.default]
    ***REMOVED***,
    filePath: this.resourcePath
  ***REMOVED***).then(jsCode => {
    if (!babel) return jsCode;
    return (0, _core.transformAsync)(jsCode, babelOptions).then(({
      code
    ***REMOVED***) => code);
  ***REMOVED***).then(result => callback(null, result)).catch(err => callback(err));

  if (previousExport) {
    readSvg().then(tranformSvg);
  ***REMOVED*** else {
    tranformSvg(source);
  ***REMOVED***
***REMOVED***

var _default = svgrLoader;
exports.default = _default;
