// jsdom/lib/jsdom/living/websockets/WebSocket-impl-browser.js
/* eslint-env browser */

"use strict";

const DOMException = require("domexception/webidl2js-wrapper");
const { parseURL, serializeURL, serializeURLOrigin ***REMOVED*** = require("whatwg-url");

const { setupForSimpleEventAccessors ***REMOVED*** = require("../helpers/create-event-accessor");
const { fireAnEvent ***REMOVED*** = require("../helpers/events");

const EventTargetImpl = require("../events/EventTarget-impl").implementation;

const idlUtils = require("../generated/utils");
const Blob = require("../generated/Blob");
const CloseEvent = require("../generated/CloseEvent");
const MessageEvent = require("../generated/MessageEvent");

const productions = {
  // https://tools.ietf.org/html/rfc7230#section-3.2.6
  token: /^[!#$%&'*+\-.^_`|~\dA-Za-z]+$/
***REMOVED***;

// https://tools.ietf.org/html/rfc6455#section-4.3
// See Sec-WebSocket-Protocol-Client, which is for the syntax of an entire header value. This function checks if a
// single header conforms to the rules.
function verifySecWebSocketProtocol(str) {
  return productions.token.test(str);
***REMOVED***

const openSockets = new WeakMap();

class WebSocketImpl extends EventTargetImpl {
  constructor(constructorArgs, privateData) {
    super([], privateData);
    const { window ***REMOVED*** = privateData;
    this._ownerDocument = idlUtils.implForWrapper(window._document);

    const url = constructorArgs[0];
    let protocols = constructorArgs[1] !== undefined ? constructorArgs[1] : [];

    const urlRecord = parseURL(url);
    if (urlRecord === null) {
      throw DOMException.create(this._globalObject, [`The URL '${url***REMOVED***' is invalid.`, "SyntaxError"]);
    ***REMOVED***
    if (urlRecord.scheme !== "ws" && urlRecord.scheme !== "wss") {
      throw DOMException.create(this._globalObject, [
        `The URL's scheme must be either 'ws' or 'wss'. '${urlRecord.scheme***REMOVED***' is not allowed.`,
        "SyntaxError"
      ]);
    ***REMOVED***
    if (urlRecord.fragment !== null) {
      throw DOMException.create(this._globalObject, [
        `The URL contains a fragment identifier ('${urlRecord.fragment***REMOVED***'). Fragment identifiers ` +
        "are not allowed in WebSocket URLs.",
        "SyntaxError"
      ]);
    ***REMOVED***

    if (typeof protocols === "string") {
      protocols = [protocols];
    ***REMOVED***
    const protocolSet = new Set();
    for (const protocol of protocols) {
      if (!verifySecWebSocketProtocol(protocol)) {
        throw DOMException.create(this._globalObject, [`The subprotocol '${protocol***REMOVED***' is invalid.`, "SyntaxError"]);
      ***REMOVED***
      const lowered = protocol.toLowerCase();
      if (protocolSet.has(lowered)) {
        throw DOMException.create(this._globalObject, [
          `The subprotocol '${protocol***REMOVED***' is duplicated.`,
          "SyntaxError"
        ]);
      ***REMOVED***
      protocolSet.add(lowered);
    ***REMOVED***

    this._urlRecord = urlRecord;
    this.url = serializeURL(urlRecord);

    this._ws = new WebSocket(this.url, protocols);

    this._ws.onopen = () => {
      fireAnEvent("open", this);
    ***REMOVED***;
    this._ws.onerror = () => {
      fireAnEvent("error", this);
    ***REMOVED***;
    this._ws.onclose = event => {
      fireAnEvent("close", this, CloseEvent, {
        wasClean: event.wasClean,
        code: event.code,
        reason: event.reason
      ***REMOVED***);
    ***REMOVED***;
    this._ws.onmessage = event => {
      fireAnEvent("message", this, MessageEvent, {
        data: event.data,
        origin: serializeURLOrigin(this._urlRecord)
      ***REMOVED***);
    ***REMOVED***;

    let openSocketsForWindow = openSockets.get(window._globalProxy);
    if (openSocketsForWindow === undefined) {
      openSocketsForWindow = new Set();
      openSockets.set(window._globalProxy, openSocketsForWindow);
    ***REMOVED***
    openSocketsForWindow.add(this);
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/web-sockets.html#make-disappear
  _makeDisappear() {
    this._eventListeners = Object.create(null);
    this._ws.close(1001);
  ***REMOVED***

  static cleanUpWindow(window) {
    const openSocketsForWindow = openSockets.get(window._globalProxy);
    if (openSocketsForWindow !== undefined) {
      for (const ws of openSocketsForWindow) {
        ws._makeDisappear();
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  get readyState() {
    return this._ws.readyState;
  ***REMOVED***

  get bufferedAmount() {
    return this._ws.bufferedAmount;
  ***REMOVED***

  get extensions() {
    return this._ws.extensions;
  ***REMOVED***

  get protocol() {
    return this._ws.protocol;
  ***REMOVED***

  close(code = undefined, reason = undefined) {
    if (code !== undefined && code !== 1000 && !(code >= 3000 && code <= 4999)) {
      throw DOMException.create(this._globalObject, [
        `The code must be either 1000, or between 3000 and 4999. ${code***REMOVED*** is neither.`,
        "InvalidAccessError"
      ]);
    ***REMOVED***
    if (reason !== undefined && Buffer.byteLength(reason, "utf8") > 123) {
      throw DOMException.create(this._globalObject, [
        "The message must not be greater than 123 bytes.",
        "SyntaxError"
      ]);
    ***REMOVED***
    return this._ws.close(code, reason);
  ***REMOVED***

  get binaryType() {
    return this._ws.binaryType;
  ***REMOVED***

  set binaryType(val) {
    this._ws.binaryType = val;
  ***REMOVED***

  send(data) {
    if (Blob.isImpl(data)) {
      data = data._buffer;
    ***REMOVED***
    this._ws.send(data);
  ***REMOVED***
***REMOVED***

setupForSimpleEventAccessors(WebSocketImpl.prototype, ["open", "message", "error", "close"]);

exports.implementation = WebSocketImpl;
