// jsdom/lib/jsdom/living/websockets/WebSocket-impl.js
"use strict";

const nodeURL = require("url");

const DOMException = require("domexception/webidl2js-wrapper");
const { parseURL, serializeURL, serializeURLOrigin ***REMOVED*** = require("whatwg-url");
const WebSocket = require("ws");

const { setupForSimpleEventAccessors ***REMOVED*** = require("../helpers/create-event-accessor");
const { fireAnEvent ***REMOVED*** = require("../helpers/events");
const { isArrayBuffer ***REMOVED*** = require("../generated/utils");
const { copyToArrayBufferInNewRealm ***REMOVED*** = require("../helpers/binary-data");

const EventTargetImpl = require("../events/EventTarget-impl").implementation;

const idlUtils = require("../generated/utils");
const Blob = require("../generated/Blob");
const CloseEvent = require("../generated/CloseEvent");
const MessageEvent = require("../generated/MessageEvent");

const CONNECTING = 0;
const OPEN = 1;
const CLOSING = 2;
const CLOSED = 3;

const productions = {
  // https://tools.ietf.org/html/rfc7230#section-3.2.6
  token: /^[!#$%&'*+\-.^_`|~\dA-Za-z]+$/
***REMOVED***;

const readyStateWSToDOM = [];
readyStateWSToDOM[WebSocket.CONNECTING] = CONNECTING;
readyStateWSToDOM[WebSocket.OPEN] = OPEN;
readyStateWSToDOM[WebSocket.CLOSING] = CLOSING;
readyStateWSToDOM[WebSocket.CLOSED] = CLOSED;

// https://tools.ietf.org/html/rfc6455#section-4.3
// See Sec-WebSocket-Protocol-Client, which is for the syntax of an entire header value. This function checks if a
// single header conforms to the rules.
function verifySecWebSocketProtocol(str) {
  return productions.token.test(str);
***REMOVED***

class PromiseQueues extends WeakMap {
  get(window) {
    const cur = super.get(window);
    return cur !== undefined ? cur : Promise.resolve();
  ***REMOVED***
***REMOVED***

const openSockets = new WeakMap();
const openingQueues = new PromiseQueues();

class WebSocketImpl extends EventTargetImpl {
  constructor(globalObject, args, privateData) {
    super(globalObject, args, privateData);

    this._ownerDocument = idlUtils.implForWrapper(globalObject._document);

    const url = args[0];
    let protocols = args[1] !== undefined ? args[1] : [];

    const urlRecord = parseURL(url);
    if (urlRecord === null) {
      throw DOMException.create(this._globalObject, [`The URL '${url***REMOVED***' is invalid.`, "SyntaxError"]);
    ***REMOVED***
    if (urlRecord.scheme !== "ws" && urlRecord.scheme !== "wss") {
      throw DOMException.create(this._globalObject, [
        `The URL's scheme must be either 'ws' or 'wss'. '${urlRecord.scheme***REMOVED***' is not allowed.`,
        "SyntaxError"
      ]);
    ***REMOVED***
    if (urlRecord.fragment !== null) {
      throw DOMException.create(this._globalObject, [
        `The URL contains a fragment identifier ('${urlRecord.fragment***REMOVED***'). Fragment identifiers ` +
        "are not allowed in WebSocket URLs.",
        "SyntaxError"
      ]);
    ***REMOVED***

    if (typeof protocols === "string") {
      protocols = [protocols];
    ***REMOVED***
    const protocolSet = new Set();
    for (const protocol of protocols) {
      if (!verifySecWebSocketProtocol(protocol)) {
        throw DOMException.create(this._globalObject, [`The subprotocol '${protocol***REMOVED***' is invalid.`, "SyntaxError"]);
      ***REMOVED***
      const lowered = protocol.toLowerCase();
      if (protocolSet.has(lowered)) {
        throw DOMException.create(this._globalObject, [
          `The subprotocol '${protocol***REMOVED***' is duplicated.`,
          "SyntaxError"
        ]);
      ***REMOVED***
      protocolSet.add(lowered);
    ***REMOVED***

    this._urlRecord = urlRecord;
    this.url = serializeURL(urlRecord);
    const nodeParsedURL = nodeURL.parse(this.url);
    this.extensions = "";

    this.binaryType = "blob";

    this._ws = null;
    // Used when this._ws has not been initialized yet.
    this._readyState = CONNECTING;
    this._requiredToFail = false;
    this.bufferedAmount = 0;
    this._sendQueue = [];

    let openSocketsForWindow = openSockets.get(globalObject._globalProxy);
    if (openSocketsForWindow === undefined) {
      openSocketsForWindow = new Set();
      openSockets.set(globalObject._globalProxy, openSocketsForWindow);
    ***REMOVED***
    openSocketsForWindow.add(this);

    openingQueues.set(this._ownerDocument, openingQueues.get(this._ownerDocument).then(() => new Promise(resolve => {
      // close() called before _ws has been initialized.
      if (this._requiredToFail) {
        resolve();
        this._readyState = CLOSED;
        this._onConnectionClosed(1006, "");
        return;
      ***REMOVED***

      this._ws = new WebSocket(this.url, protocols, {
        headers: {
          "user-agent": globalObject.navigator.userAgent,
          "cookie": this._ownerDocument._cookieJar.getCookieStringSync(nodeParsedURL, { http: true ***REMOVED***),
          "origin": globalObject._origin
        ***REMOVED***,
        rejectUnauthorized: this._ownerDocument._strictSSL
      ***REMOVED***);
      this._ws.once("open", () => {
        resolve();
        this._onConnectionEstablished();
      ***REMOVED***);
      this._ws.on("message", this._onMessageReceived.bind(this));
      this._ws.once("close", (...closeArgs) => {
        resolve();
        this._onConnectionClosed(...closeArgs);
      ***REMOVED***);
      this._ws.once("upgrade", ({ headers ***REMOVED***) => {
        if (Array.isArray(headers["set-cookie"])) {
          for (const cookie of headers["set-cookie"]) {
            this._ownerDocument._cookieJar.setCookieSync(
              cookie,
              nodeParsedURL,
              { http: true, ignoreError: true ***REMOVED***
            );
          ***REMOVED***
        ***REMOVED*** else if (headers["set-cookie"] !== undefined) {
          this._ownerDocument._cookieJar.setCookieSync(
            headers["set-cookie"],
            nodeParsedURL,
            { http: true, ignoreError: true ***REMOVED***
          );
        ***REMOVED***
      ***REMOVED***);
      this._ws.once("error", () => {
        // The exact error is passed into this callback, but it is ignored as we don't really care about it.
        resolve();
        this._requiredToFail = true;
        // Do not emit an error here, as that will be handled in _onConnectionClosed. ws always emits a close event
        // after errors.
      ***REMOVED***);
    ***REMOVED***)));
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/web-sockets.html#make-disappear
  _makeDisappear() {
    this._eventListeners = Object.create(null);
    this._close(1001);
  ***REMOVED***

  static cleanUpWindow(window) {
    const openSocketsForWindow = openSockets.get(window._globalProxy);
    if (openSocketsForWindow !== undefined) {
      for (const ws of openSocketsForWindow) {
        ws._makeDisappear();
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/web-sockets.html#feedback-from-the-protocol
  _onConnectionEstablished() {
    // readyState is a getter.
    if (this._ws.extensions !== null) {
      // Right now, ws only supports one extension, permessage-deflate, without any parameters. This algorithm may need
      // to be more sophiscated as more extenions are supported.
      this.extensions = Object.keys(this._ws.extensions).join(", ");
    ***REMOVED***
    // protocol is a getter.
    fireAnEvent("open", this);
  ***REMOVED***

  _onMessageReceived(data) {
    if (this.readyState !== OPEN) {
      return;
    ***REMOVED***
    let dataForEvent;
    if (typeof data === "string") {
      dataForEvent = data;
    ***REMOVED*** else if (this.binaryType === "arraybuffer") {
      if (isArrayBuffer(data)) {
        dataForEvent = data;
      ***REMOVED*** else if (Array.isArray(data)) {
        dataForEvent = copyToArrayBufferInNewRealm(Buffer.concat(data), this._globalObject);
      ***REMOVED*** else {
        dataForEvent = copyToArrayBufferInNewRealm(data, this._globalObject);
      ***REMOVED***
    ***REMOVED*** else { // this.binaryType === "blob"
      if (!Array.isArray(data)) {
        data = [data];
      ***REMOVED***
      dataForEvent = Blob.create(this._globalObject, [data, { type: "" ***REMOVED***]);
    ***REMOVED***
    fireAnEvent("message", this, MessageEvent, {
      data: dataForEvent,
      origin: serializeURLOrigin(this._urlRecord)
    ***REMOVED***);
  ***REMOVED***

  _onConnectionClosed(code, reason) {
    const openSocketsForWindow = openSockets.get(this._ownerDocument._defaultView);
    openSocketsForWindow.delete(this);

    const wasClean = !this._requiredToFail;
    if (this._requiredToFail) {
      fireAnEvent("error", this);
    ***REMOVED***
    fireAnEvent("close", this, CloseEvent, {
      wasClean,
      code,
      reason
    ***REMOVED***);
  ***REMOVED***

  get readyState() {
    if (this._ws !== null) {
      return readyStateWSToDOM[this._ws.readyState];
    ***REMOVED***
    return this._readyState;
  ***REMOVED***

  get protocol() {
    if (this._ws === null) {
      return "";
    ***REMOVED***
    return this._ws.protocol;
  ***REMOVED***

  close(code = undefined, reason = undefined) {
    if (code !== undefined && code !== 1000 && !(code >= 3000 && code <= 4999)) {
      throw DOMException.create(this._globalObject, [
        `The code must be either 1000, or between 3000 and 4999. ${code***REMOVED*** is neither.`,
        "InvalidAccessError"
      ]);
    ***REMOVED***
    if (reason !== undefined && Buffer.byteLength(reason, "utf8") > 123) {
      throw DOMException.create(this._globalObject, [
        "The message must not be greater than 123 bytes.",
        "SyntaxError"
      ]);
    ***REMOVED***
    this._close(code, reason);
  ***REMOVED***

  _close(code = undefined, reason = undefined) {
    if (this.readyState === CONNECTING) {
      this._requiredToFail = true;
      if (this._ws !== null) {
        this._ws.terminate();
      ***REMOVED*** else {
        this._readyState = CLOSING;
      ***REMOVED***
    ***REMOVED*** else if (this.readyState === OPEN) {
      this._ws.close(code, reason);
    ***REMOVED***
  ***REMOVED***

  send(data) {
    if (this.readyState === CONNECTING) {
      throw DOMException.create(this._globalObject, ["Still in CONNECTING state.", "InvalidStateError"]);
    ***REMOVED***
    if (this.readyState !== OPEN) {
      return;
    ***REMOVED***
    if (Blob.isImpl(data)) {
      data = data._buffer;
    ***REMOVED***
    let length;
    if (typeof data === "string") {
      length = Buffer.byteLength(data, "utf8");
    ***REMOVED*** else {
      length = data.byteLength;
    ***REMOVED***
    this.bufferedAmount += length;
    this._sendQueue.push([data, length]);
    this._scheduleSend();
  ***REMOVED***

  _actuallySend() {
    for (const [data, length] of this._sendQueue.splice(0)) {
      this._ws.send(data, { binary: typeof data !== "string" ***REMOVED***, () => {
        this.bufferedAmount -= length;
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***

  _scheduleSend() {
    if (this._dequeueScheduled) {
      return;
    ***REMOVED***
    this._dequeueScheduled = true;
    process.nextTick(() => {
      this._dequeueScheduled = false;
      this._actuallySend();
    ***REMOVED***);
  ***REMOVED***
***REMOVED***

setupForSimpleEventAccessors(WebSocketImpl.prototype, ["open", "message", "error", "close"]);

exports.implementation = WebSocketImpl;
