// postcss-js/objectifier.js
let camelcase = require('camelcase-css')

let UNITLESS = {
  boxFlex: true,
  boxFlexGroup: true,
  columnCount: true,
  flex: true,
  flexGrow: true,
  flexPositive: true,
  flexShrink: true,
  flexNegative: true,
  fontWeight: true,
  lineClamp: true,
  lineHeight: true,
  opacity: true,
  order: true,
  orphans: true,
  tabSize: true,
  widows: true,
  zIndex: true,
  zoom: true,
  fillOpacity: true,
  strokeDashoffset: true,
  strokeOpacity: true,
  strokeWidth: true
***REMOVED***

function atRule(node) {
  if (typeof node.nodes === 'undefined') {
    return true
  ***REMOVED*** else {
    return process(node)
  ***REMOVED***
***REMOVED***

function process(node) {
  let name
  let result = {***REMOVED***

  node.each(child => {
    if (child.type === 'atrule') {
      name = '@' + child.name
      if (child.params) name += ' ' + child.params
      if (typeof result[name] === 'undefined') {
        result[name] = atRule(child)
      ***REMOVED*** else if (Array.isArray(result[name])) {
        result[name].push(atRule(child))
      ***REMOVED*** else {
        result[name] = [result[name], atRule(child)]
      ***REMOVED***
    ***REMOVED*** else if (child.type === 'rule') {
      let body = process(child)
      if (result[child.selector]) {
        for (let i in body) {
          result[child.selector][i] = body[i]
        ***REMOVED***
      ***REMOVED*** else {
        result[child.selector] = body
      ***REMOVED***
    ***REMOVED*** else if (child.type === 'decl') {
      if (child.prop[0] === '-' && child.prop[1] === '-') {
        name = child.prop
      ***REMOVED*** else if (child.parent && child.parent.selector === ':export') {
        name = child.prop
      ***REMOVED*** else {
        name = camelcase(child.prop)
      ***REMOVED***
      let value = child.value
      if (!isNaN(child.value) && UNITLESS[name]) {
        value = parseFloat(child.value)
      ***REMOVED***
      if (child.important) value += ' !important'
      if (typeof result[name] === 'undefined') {
        result[name] = value
      ***REMOVED*** else if (Array.isArray(result[name])) {
        result[name].push(value)
      ***REMOVED*** else {
        result[name] = [result[name], value]
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***)
  return result
***REMOVED***

module.exports = process
