// @babel/helper-module-imports/lib/import-injector.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var _assert = require("assert");
var _t = require("@babel/types");
var _importBuilder = require("./import-builder.js");
var _isModule = require("./is-module.js");
const {
  identifier,
  importSpecifier,
  numericLiteral,
  sequenceExpression,
  isImportDeclaration
***REMOVED*** = _t;
class ImportInjector {
  constructor(path, importedSource, opts) {
    this._defaultOpts = {
      importedSource: null,
      importedType: "commonjs",
      importedInterop: "babel",
      importingInterop: "babel",
      ensureLiveReference: false,
      ensureNoContext: false,
      importPosition: "before"
    ***REMOVED***;
    const programPath = path.find(p => p.isProgram());
    this._programPath = programPath;
    this._programScope = programPath.scope;
    this._hub = programPath.hub;
    this._defaultOpts = this._applyDefaults(importedSource, opts, true);
  ***REMOVED***
  addDefault(importedSourceIn, opts) {
    return this.addNamed("default", importedSourceIn, opts);
  ***REMOVED***
  addNamed(importName, importedSourceIn, opts) {
    _assert(typeof importName === "string");
    return this._generateImport(this._applyDefaults(importedSourceIn, opts), importName);
  ***REMOVED***
  addNamespace(importedSourceIn, opts) {
    return this._generateImport(this._applyDefaults(importedSourceIn, opts), null);
  ***REMOVED***
  addSideEffect(importedSourceIn, opts) {
    return this._generateImport(this._applyDefaults(importedSourceIn, opts), void 0);
  ***REMOVED***
  _applyDefaults(importedSource, opts, isInit = false) {
    let newOpts;
    if (typeof importedSource === "string") {
      newOpts = Object.assign({***REMOVED***, this._defaultOpts, {
        importedSource
      ***REMOVED***, opts);
    ***REMOVED*** else {
      _assert(!opts, "Unexpected secondary arguments.");
      newOpts = Object.assign({***REMOVED***, this._defaultOpts, importedSource);
    ***REMOVED***
    if (!isInit && opts) {
      if (opts.nameHint !== undefined) newOpts.nameHint = opts.nameHint;
      if (opts.blockHoist !== undefined) newOpts.blockHoist = opts.blockHoist;
    ***REMOVED***
    return newOpts;
  ***REMOVED***
  _generateImport(opts, importName) {
    const isDefault = importName === "default";
    const isNamed = !!importName && !isDefault;
    const isNamespace = importName === null;
    const {
      importedSource,
      importedType,
      importedInterop,
      importingInterop,
      ensureLiveReference,
      ensureNoContext,
      nameHint,
      importPosition,
      blockHoist
    ***REMOVED*** = opts;
    let name = nameHint || importName;
    const isMod = (0, _isModule.default)(this._programPath);
    const isModuleForNode = isMod && importingInterop === "node";
    const isModuleForBabel = isMod && importingInterop === "babel";
    if (importPosition === "after" && !isMod) {
      throw new Error(`"importPosition": "after" is only supported in modules`);
    ***REMOVED***
    const builder = new _importBuilder.default(importedSource, this._programScope, this._hub);
    if (importedType === "es6") {
      if (!isModuleForNode && !isModuleForBabel) {
        throw new Error("Cannot import an ES6 module from CommonJS");
      ***REMOVED***
      builder.import();
      if (isNamespace) {
        builder.namespace(nameHint || importedSource);
      ***REMOVED*** else if (isDefault || isNamed) {
        builder.named(name, importName);
      ***REMOVED***
    ***REMOVED*** else if (importedType !== "commonjs") {
      throw new Error(`Unexpected interopType "${importedType***REMOVED***"`);
    ***REMOVED*** else if (importedInterop === "babel") {
      if (isModuleForNode) {
        name = name !== "default" ? name : importedSource;
        const es6Default = `${importedSource***REMOVED***$es6Default`;
        builder.import();
        if (isNamespace) {
          builder.default(es6Default).var(name || importedSource).wildcardInterop();
        ***REMOVED*** else if (isDefault) {
          if (ensureLiveReference) {
            builder.default(es6Default).var(name || importedSource).defaultInterop().read("default");
          ***REMOVED*** else {
            builder.default(es6Default).var(name).defaultInterop().prop(importName);
          ***REMOVED***
        ***REMOVED*** else if (isNamed) {
          builder.default(es6Default).read(importName);
        ***REMOVED***
      ***REMOVED*** else if (isModuleForBabel) {
        builder.import();
        if (isNamespace) {
          builder.namespace(name || importedSource);
        ***REMOVED*** else if (isDefault || isNamed) {
          builder.named(name, importName);
        ***REMOVED***
      ***REMOVED*** else {
        builder.require();
        if (isNamespace) {
          builder.var(name || importedSource).wildcardInterop();
        ***REMOVED*** else if ((isDefault || isNamed) && ensureLiveReference) {
          if (isDefault) {
            name = name !== "default" ? name : importedSource;
            builder.var(name).read(importName);
            builder.defaultInterop();
          ***REMOVED*** else {
            builder.var(importedSource).read(importName);
          ***REMOVED***
        ***REMOVED*** else if (isDefault) {
          builder.var(name).defaultInterop().prop(importName);
        ***REMOVED*** else if (isNamed) {
          builder.var(name).prop(importName);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED*** else if (importedInterop === "compiled") {
      if (isModuleForNode) {
        builder.import();
        if (isNamespace) {
          builder.default(name || importedSource);
        ***REMOVED*** else if (isDefault || isNamed) {
          builder.default(importedSource).read(name);
        ***REMOVED***
      ***REMOVED*** else if (isModuleForBabel) {
        builder.import();
        if (isNamespace) {
          builder.namespace(name || importedSource);
        ***REMOVED*** else if (isDefault || isNamed) {
          builder.named(name, importName);
        ***REMOVED***
      ***REMOVED*** else {
        builder.require();
        if (isNamespace) {
          builder.var(name || importedSource);
        ***REMOVED*** else if (isDefault || isNamed) {
          if (ensureLiveReference) {
            builder.var(importedSource).read(name);
          ***REMOVED*** else {
            builder.prop(importName).var(name);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED*** else if (importedInterop === "uncompiled") {
      if (isDefault && ensureLiveReference) {
        throw new Error("No live reference for commonjs default");
      ***REMOVED***
      if (isModuleForNode) {
        builder.import();
        if (isNamespace) {
          builder.default(name || importedSource);
        ***REMOVED*** else if (isDefault) {
          builder.default(name);
        ***REMOVED*** else if (isNamed) {
          builder.default(importedSource).read(name);
        ***REMOVED***
      ***REMOVED*** else if (isModuleForBabel) {
        builder.import();
        if (isNamespace) {
          builder.default(name || importedSource);
        ***REMOVED*** else if (isDefault) {
          builder.default(name);
        ***REMOVED*** else if (isNamed) {
          builder.named(name, importName);
        ***REMOVED***
      ***REMOVED*** else {
        builder.require();
        if (isNamespace) {
          builder.var(name || importedSource);
        ***REMOVED*** else if (isDefault) {
          builder.var(name);
        ***REMOVED*** else if (isNamed) {
          if (ensureLiveReference) {
            builder.var(importedSource).read(name);
          ***REMOVED*** else {
            builder.var(name).prop(importName);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED*** else {
      throw new Error(`Unknown importedInterop "${importedInterop***REMOVED***".`);
    ***REMOVED***
    const {
      statements,
      resultName
    ***REMOVED*** = builder.done();
    this._insertStatements(statements, importPosition, blockHoist);
    if ((isDefault || isNamed) && ensureNoContext && resultName.type !== "Identifier") {
      return sequenceExpression([numericLiteral(0), resultName]);
    ***REMOVED***
    return resultName;
  ***REMOVED***
  _insertStatements(statements, importPosition = "before", blockHoist = 3) {
    if (importPosition === "after") {
      if (this._insertStatementsAfter(statements)) return;
    ***REMOVED*** else {
      if (this._insertStatementsBefore(statements, blockHoist)) return;
    ***REMOVED***
    this._programPath.unshiftContainer("body", statements);
  ***REMOVED***
  _insertStatementsBefore(statements, blockHoist) {
    if (statements.length === 1 && isImportDeclaration(statements[0]) && isValueImport(statements[0])) {
      const firstImportDecl = this._programPath.get("body").find(p => {
        return p.isImportDeclaration() && isValueImport(p.node);
      ***REMOVED***);
      if ((firstImportDecl == null ? void 0 : firstImportDecl.node.source.value) === statements[0].source.value && maybeAppendImportSpecifiers(firstImportDecl.node, statements[0])) {
        return true;
      ***REMOVED***
    ***REMOVED***
    statements.forEach(node => {
      node._blockHoist = blockHoist;
    ***REMOVED***);
    const targetPath = this._programPath.get("body").find(p => {
      const val = p.node._blockHoist;
      return Number.isFinite(val) && val < 4;
    ***REMOVED***);
    if (targetPath) {
      targetPath.insertBefore(statements);
      return true;
    ***REMOVED***
    return false;
  ***REMOVED***
  _insertStatementsAfter(statements) {
    const statementsSet = new Set(statements);
    const importDeclarations = new Map();
    for (const statement of statements) {
      if (isImportDeclaration(statement) && isValueImport(statement)) {
        const source = statement.source.value;
        if (!importDeclarations.has(source)) importDeclarations.set(source, []);
        importDeclarations.get(source).push(statement);
      ***REMOVED***
    ***REMOVED***
    let lastImportPath = null;
    for (const bodyStmt of this._programPath.get("body")) {
      if (bodyStmt.isImportDeclaration() && isValueImport(bodyStmt.node)) {
        lastImportPath = bodyStmt;
        const source = bodyStmt.node.source.value;
        const newImports = importDeclarations.get(source);
        if (!newImports) continue;
        for (const decl of newImports) {
          if (!statementsSet.has(decl)) continue;
          if (maybeAppendImportSpecifiers(bodyStmt.node, decl)) {
            statementsSet.delete(decl);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
    if (statementsSet.size === 0) return true;
    if (lastImportPath) lastImportPath.insertAfter(Array.from(statementsSet));
    return !!lastImportPath;
  ***REMOVED***
***REMOVED***
exports.default = ImportInjector;
function isValueImport(node) {
  return node.importKind !== "type" && node.importKind !== "typeof";
***REMOVED***
function hasNamespaceImport(node) {
  return node.specifiers.length === 1 && node.specifiers[0].type === "ImportNamespaceSpecifier" || node.specifiers.length === 2 && node.specifiers[1].type === "ImportNamespaceSpecifier";
***REMOVED***
function hasDefaultImport(node) {
  return node.specifiers.length > 0 && node.specifiers[0].type === "ImportDefaultSpecifier";
***REMOVED***
function maybeAppendImportSpecifiers(target, source) {
  if (!target.specifiers.length) {
    target.specifiers = source.specifiers;
    return true;
  ***REMOVED***
  if (!source.specifiers.length) return true;
  if (hasNamespaceImport(target) || hasNamespaceImport(source)) return false;
  if (hasDefaultImport(source)) {
    if (hasDefaultImport(target)) {
      source.specifiers[0] = importSpecifier(source.specifiers[0].local, identifier("default"));
    ***REMOVED*** else {
      target.specifiers.unshift(source.specifiers.shift());
    ***REMOVED***
  ***REMOVED***
  target.specifiers.push(...source.specifiers);
  return true;
***REMOVED***
