// postcss-load-config/src/index.js
'use strict'

const resolve = require('path').resolve
const url = require('url')

const config = require('lilconfig')
const yaml = require('yaml')

const loadOptions = require('./options.js')
const loadPlugins = require('./plugins.js')

/* istanbul ignore next */
const interopRequireDefault = (obj) => obj && obj.__esModule ? obj : { default: obj ***REMOVED***

/**
 * Process the result from cosmiconfig
 *
 * @param  {Object***REMOVED*** ctx Config Context
 * @param  {Object***REMOVED*** result Cosmiconfig result
 *
 * @return {Object***REMOVED*** PostCSS Config
 */
const processResult = (ctx, result) => {
  const file = result.filepath || ''
  let config = interopRequireDefault(result.config).default || {***REMOVED***

  if (typeof config === 'function') {
    config = config(ctx)
  ***REMOVED*** else {
    config = Object.assign({***REMOVED***, config, ctx)
  ***REMOVED***

  if (!config.plugins) {
    config.plugins = []
  ***REMOVED***

  return {
    plugins: loadPlugins(config, file),
    options: loadOptions(config, file),
    file
  ***REMOVED***
***REMOVED***

/**
 * Builds the Config Context
 *
 * @param  {Object***REMOVED*** ctx Config Context
 *
 * @return {Object***REMOVED*** Config Context
 */
const createContext = (ctx) => {
  /**
   * @type {Object***REMOVED***
   *
   * @prop {String***REMOVED*** cwd=process.cwd() Config search start location
   * @prop {String***REMOVED*** env=process.env.NODE_ENV Config Enviroment, will be set to `development` by `postcss-load-config` if `process.env.NODE_ENV` is `undefined`
   */
  ctx = Object.assign({
    cwd: process.cwd(),
    env: process.env.NODE_ENV
  ***REMOVED***, ctx)

  if (!ctx.env) {
    process.env.NODE_ENV = 'development'
  ***REMOVED***

  return ctx
***REMOVED***

const importDefault = async filepath => {
  const module = await import(url.pathToFileURL(filepath).href)
  return module.default
***REMOVED***

const addTypeScriptLoader = (options = {***REMOVED***, loader) => {
  const moduleName = 'postcss'

  return {
    ...options,
    searchPlaces: [
      ...(options.searchPlaces || []),
      'package.json',
      `.${moduleName***REMOVED***rc`,
      `.${moduleName***REMOVED***rc.json`,
      `.${moduleName***REMOVED***rc.yaml`,
      `.${moduleName***REMOVED***rc.yml`,
      `.${moduleName***REMOVED***rc.ts`,
      `.${moduleName***REMOVED***rc.cts`,
      `.${moduleName***REMOVED***rc.js`,
      `.${moduleName***REMOVED***rc.cjs`,
      `.${moduleName***REMOVED***rc.mjs`,
      `${moduleName***REMOVED***.config.ts`,
      `${moduleName***REMOVED***.config.cts`,
      `${moduleName***REMOVED***.config.js`,
      `${moduleName***REMOVED***.config.cjs`,
      `${moduleName***REMOVED***.config.mjs`
    ],
    loaders: {
      ...options.loaders,
      '.yaml': (filepath, content) => yaml.parse(content),
      '.yml': (filepath, content) => yaml.parse(content),
      '.js': importDefault,
      '.cjs': importDefault,
      '.mjs': importDefault,
      '.ts': loader,
      '.cts': loader
    ***REMOVED***
  ***REMOVED***
***REMOVED***

const withTypeScriptLoader = (rcFunc) => {
  return (ctx, path, options) => {
    return rcFunc(ctx, path, addTypeScriptLoader(options, (configFile) => {
      let registerer = { enabled () {***REMOVED*** ***REMOVED***

      try {
        // Register TypeScript compiler instance
        registerer = require('ts-node').register({
          // transpile to cjs even if compilerOptions.module in tsconfig is not Node16/NodeNext.
          moduleTypes: { '**/*.cts': 'cjs' ***REMOVED***
        ***REMOVED***)

        return require(configFile)
      ***REMOVED*** catch (err) {
        if (err.code === 'MODULE_NOT_FOUND') {
          throw new Error(
            `'ts-node' is required for the TypeScript configuration files. Make sure it is installed\nError: ${err.message***REMOVED***`
          )
        ***REMOVED***

        throw err
      ***REMOVED*** finally {
        registerer.enabled(false)
      ***REMOVED***
    ***REMOVED***))
  ***REMOVED***
***REMOVED***

/**
 * Load Config
 *
 * @method rc
 *
 * @param  {Object***REMOVED*** ctx Config Context
 * @param  {String***REMOVED*** path Config Path
 * @param  {Object***REMOVED*** options Config Options
 *
 * @return {Promise***REMOVED*** config PostCSS Config
 */
const rc = withTypeScriptLoader((ctx, path, options) => {
  /**
   * @type {Object***REMOVED*** The full Config Context
   */
  ctx = createContext(ctx)

  /**
   * @type {String***REMOVED*** `process.cwd()`
   */
  path = path ? resolve(path) : process.cwd()

  return config.lilconfig('postcss', options)
    .search(path)
    .then((result) => {
      if (!result) {
        throw new Error(`No PostCSS Config found in: ${path***REMOVED***`)
      ***REMOVED***

      return processResult(ctx, result)
    ***REMOVED***)
***REMOVED***)

/**
 * Autoload Config for PostCSS
 *
 * @author Michael Ciniawsky @michael-ciniawsky <michael.ciniawsky@gmail.com>
 * @license MIT
 *
 * @module postcss-load-config
 * @version 2.1.0
 *
 * @requires comsiconfig
 * @requires ./options
 * @requires ./plugins
 */
module.exports = rc
