// webpack/lib/logging/Logger.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const LogType = Object.freeze({
	error: /** @type {"error"***REMOVED*** */ ("error"), // message, c style arguments
	warn: /** @type {"warn"***REMOVED*** */ ("warn"), // message, c style arguments
	info: /** @type {"info"***REMOVED*** */ ("info"), // message, c style arguments
	log: /** @type {"log"***REMOVED*** */ ("log"), // message, c style arguments
	debug: /** @type {"debug"***REMOVED*** */ ("debug"), // message, c style arguments

	trace: /** @type {"trace"***REMOVED*** */ ("trace"), // no arguments

	group: /** @type {"group"***REMOVED*** */ ("group"), // [label]
	groupCollapsed: /** @type {"groupCollapsed"***REMOVED*** */ ("groupCollapsed"), // [label]
	groupEnd: /** @type {"groupEnd"***REMOVED*** */ ("groupEnd"), // [label]

	profile: /** @type {"profile"***REMOVED*** */ ("profile"), // [profileName]
	profileEnd: /** @type {"profileEnd"***REMOVED*** */ ("profileEnd"), // [profileName]

	time: /** @type {"time"***REMOVED*** */ ("time"), // name, time as [seconds, nanoseconds]

	clear: /** @type {"clear"***REMOVED*** */ ("clear"), // no arguments
	status: /** @type {"status"***REMOVED*** */ ("status") // message, arguments
***REMOVED***);

module.exports.LogType = LogType;

/** @typedef {typeof LogType[keyof typeof LogType]***REMOVED*** LogTypeEnum */

const LOG_SYMBOL = Symbol("webpack logger raw log method");
const TIMERS_SYMBOL = Symbol("webpack logger times");
const TIMERS_AGGREGATES_SYMBOL = Symbol("webpack logger aggregated times");

class WebpackLogger {
	/**
	 * @param {(type: LogTypeEnum, args?: EXPECTED_ANY[]) => void***REMOVED*** log log function
	 * @param {(name: string | (() => string)) => WebpackLogger***REMOVED*** getChildLogger function to create child logger
	 */
	constructor(log, getChildLogger) {
		this[LOG_SYMBOL] = log;
		this.getChildLogger = getChildLogger;
	***REMOVED***

	/**
	 * @param {...EXPECTED_ANY***REMOVED*** args args
	 */
	error(...args) {
		this[LOG_SYMBOL](LogType.error, args);
	***REMOVED***

	/**
	 * @param {...EXPECTED_ANY***REMOVED*** args args
	 */
	warn(...args) {
		this[LOG_SYMBOL](LogType.warn, args);
	***REMOVED***

	/**
	 * @param {...EXPECTED_ANY***REMOVED*** args args
	 */
	info(...args) {
		this[LOG_SYMBOL](LogType.info, args);
	***REMOVED***

	/**
	 * @param {...EXPECTED_ANY***REMOVED*** args args
	 */
	log(...args) {
		this[LOG_SYMBOL](LogType.log, args);
	***REMOVED***

	/**
	 * @param {...EXPECTED_ANY***REMOVED*** args args
	 */
	debug(...args) {
		this[LOG_SYMBOL](LogType.debug, args);
	***REMOVED***

	/**
	 * @param {EXPECTED_ANY***REMOVED*** assertion assertion
	 * @param {...EXPECTED_ANY***REMOVED*** args args
	 */
	assert(assertion, ...args) {
		if (!assertion) {
			this[LOG_SYMBOL](LogType.error, args);
		***REMOVED***
	***REMOVED***

	trace() {
		this[LOG_SYMBOL](LogType.trace, ["Trace"]);
	***REMOVED***

	clear() {
		this[LOG_SYMBOL](LogType.clear);
	***REMOVED***

	/**
	 * @param {...EXPECTED_ANY***REMOVED*** args args
	 */
	status(...args) {
		this[LOG_SYMBOL](LogType.status, args);
	***REMOVED***

	/**
	 * @param {...EXPECTED_ANY***REMOVED*** args args
	 */
	group(...args) {
		this[LOG_SYMBOL](LogType.group, args);
	***REMOVED***

	/**
	 * @param {...EXPECTED_ANY***REMOVED*** args args
	 */
	groupCollapsed(...args) {
		this[LOG_SYMBOL](LogType.groupCollapsed, args);
	***REMOVED***

	groupEnd() {
		this[LOG_SYMBOL](LogType.groupEnd);
	***REMOVED***

	/**
	 * @param {string=***REMOVED*** label label
	 */
	profile(label) {
		this[LOG_SYMBOL](LogType.profile, [label]);
	***REMOVED***

	/**
	 * @param {string=***REMOVED*** label label
	 */
	profileEnd(label) {
		this[LOG_SYMBOL](LogType.profileEnd, [label]);
	***REMOVED***

	/**
	 * @param {string***REMOVED*** label label
	 */
	time(label) {
		/** @type {Map<string | undefined, [number, number]>***REMOVED*** */
		this[TIMERS_SYMBOL] = this[TIMERS_SYMBOL] || new Map();
		this[TIMERS_SYMBOL].set(label, process.hrtime());
	***REMOVED***

	/**
	 * @param {string=***REMOVED*** label label
	 */
	timeLog(label) {
		const prev = this[TIMERS_SYMBOL] && this[TIMERS_SYMBOL].get(label);
		if (!prev) {
			throw new Error(`No such label '${label***REMOVED***' for WebpackLogger.timeLog()`);
		***REMOVED***
		const time = process.hrtime(prev);
		this[LOG_SYMBOL](LogType.time, [label, ...time]);
	***REMOVED***

	/**
	 * @param {string=***REMOVED*** label label
	 */
	timeEnd(label) {
		const prev = this[TIMERS_SYMBOL] && this[TIMERS_SYMBOL].get(label);
		if (!prev) {
			throw new Error(`No such label '${label***REMOVED***' for WebpackLogger.timeEnd()`);
		***REMOVED***
		const time = process.hrtime(prev);
		/** @type {Map<string | undefined, [number, number]>***REMOVED*** */
		(this[TIMERS_SYMBOL]).delete(label);
		this[LOG_SYMBOL](LogType.time, [label, ...time]);
	***REMOVED***

	/**
	 * @param {string=***REMOVED*** label label
	 */
	timeAggregate(label) {
		const prev = this[TIMERS_SYMBOL] && this[TIMERS_SYMBOL].get(label);
		if (!prev) {
			throw new Error(
				`No such label '${label***REMOVED***' for WebpackLogger.timeAggregate()`
			);
		***REMOVED***
		const time = process.hrtime(prev);
		/** @type {Map<string | undefined, [number, number]>***REMOVED*** */
		(this[TIMERS_SYMBOL]).delete(label);
		/** @type {Map<string | undefined, [number, number]>***REMOVED*** */
		this[TIMERS_AGGREGATES_SYMBOL] =
			this[TIMERS_AGGREGATES_SYMBOL] || new Map();
		const current = this[TIMERS_AGGREGATES_SYMBOL].get(label);
		if (current !== undefined) {
			if (time[1] + current[1] > 1e9) {
				time[0] += current[0] + 1;
				time[1] = time[1] - 1e9 + current[1];
			***REMOVED*** else {
				time[0] += current[0];
				time[1] += current[1];
			***REMOVED***
		***REMOVED***
		this[TIMERS_AGGREGATES_SYMBOL].set(label, time);
	***REMOVED***

	/**
	 * @param {string=***REMOVED*** label label
	 */
	timeAggregateEnd(label) {
		if (this[TIMERS_AGGREGATES_SYMBOL] === undefined) return;
		const time = this[TIMERS_AGGREGATES_SYMBOL].get(label);
		if (time === undefined) return;
		this[TIMERS_AGGREGATES_SYMBOL].delete(label);
		this[LOG_SYMBOL](LogType.time, [label, ...time]);
	***REMOVED***
***REMOVED***

module.exports.Logger = WebpackLogger;
