// autoprefixer/lib/browsers.js
let browserslist = require('browserslist')
let { agents ***REMOVED*** = require('caniuse-lite/dist/unpacker/agents')

let utils = require('./utils')

class Browsers {
  constructor(data, requirements, options, browserslistOpts) {
    this.data = data
    this.options = options || {***REMOVED***
    this.browserslistOpts = browserslistOpts || {***REMOVED***
    this.selected = this.parse(requirements)
  ***REMOVED***

  /**
   * Return all prefixes for default browser data
   */
  static prefixes() {
    if (this.prefixesCache) {
      return this.prefixesCache
    ***REMOVED***

    this.prefixesCache = []
    for (let name in agents) {
      this.prefixesCache.push(`-${agents[name].prefix***REMOVED***-`)
    ***REMOVED***

    this.prefixesCache = utils
      .uniq(this.prefixesCache)
      .sort((a, b) => b.length - a.length)

    return this.prefixesCache
  ***REMOVED***

  /**
   * Check is value contain any possible prefix
   */
  static withPrefix(value) {
    if (!this.prefixesRegexp) {
      this.prefixesRegexp = new RegExp(this.prefixes().join('|'))
    ***REMOVED***

    return this.prefixesRegexp.test(value)
  ***REMOVED***

  /**
   * Is browser is selected by requirements
   */
  isSelected(browser) {
    return this.selected.includes(browser)
  ***REMOVED***

  /**
   * Return browsers selected by requirements
   */
  parse(requirements) {
    let opts = {***REMOVED***
    for (let i in this.browserslistOpts) {
      opts[i] = this.browserslistOpts[i]
    ***REMOVED***
    opts.path = this.options.from
    return browserslist(requirements, opts)
  ***REMOVED***

  /**
   * Return prefix for selected browser
   */
  prefix(browser) {
    let [name, version] = browser.split(' ')
    let data = this.data[name]

    let prefix = data.prefix_exceptions && data.prefix_exceptions[version]
    if (!prefix) {
      prefix = data.prefix
    ***REMOVED***
    return `-${prefix***REMOVED***-`
  ***REMOVED***
***REMOVED***

module.exports = Browsers
