// autoprefixer/lib/declaration.js
let Browsers = require('./browsers')
let Prefixer = require('./prefixer')
let utils = require('./utils')

class Declaration extends Prefixer {
  /**
   * Clone and add prefixes for declaration
   */
  add(decl, prefix, prefixes, result) {
    let prefixed = this.prefixed(decl.prop, prefix)
    if (
      this.isAlready(decl, prefixed) ||
      this.otherPrefixes(decl.value, prefix)
    ) {
      return undefined
    ***REMOVED***
    return this.insert(decl, prefix, prefixes, result)
  ***REMOVED***

  /**
   * Calculate indentation to create visual cascade
   */
  calcBefore(prefixes, decl, prefix = '') {
    let max = this.maxPrefixed(prefixes, decl)
    let diff = max - utils.removeNote(prefix).length

    let before = decl.raw('before')
    if (diff > 0) {
      before += Array(diff).fill(' ').join('')
    ***REMOVED***

    return before
  ***REMOVED***

  /**
   * Always true, because we already get prefixer by property name
   */
  check(/* decl */) {
    return true
  ***REMOVED***

  /**
   * Clone and insert new declaration
   */
  insert(decl, prefix, prefixes) {
    let cloned = this.set(this.clone(decl), prefix)
    if (!cloned) return undefined

    let already = decl.parent.some(
      i => i.prop === cloned.prop && i.value === cloned.value
    )
    if (already) {
      return undefined
    ***REMOVED***

    if (this.needCascade(decl)) {
      cloned.raws.before = this.calcBefore(prefixes, decl, prefix)
    ***REMOVED***
    return decl.parent.insertBefore(decl, cloned)
  ***REMOVED***

  /**
   * Did this declaration has this prefix above
   */
  isAlready(decl, prefixed) {
    let already = this.all.group(decl).up(i => i.prop === prefixed)
    if (!already) {
      already = this.all.group(decl).down(i => i.prop === prefixed)
    ***REMOVED***
    return already
  ***REMOVED***

  /**
   * Return maximum length of possible prefixed property
   */
  maxPrefixed(prefixes, decl) {
    if (decl._autoprefixerMax) {
      return decl._autoprefixerMax
    ***REMOVED***

    let max = 0
    for (let prefix of prefixes) {
      prefix = utils.removeNote(prefix)
      if (prefix.length > max) {
        max = prefix.length
      ***REMOVED***
    ***REMOVED***
    decl._autoprefixerMax = max

    return decl._autoprefixerMax
  ***REMOVED***

  /**
   * Should we use visual cascade for prefixes
   */
  needCascade(decl) {
    if (!decl._autoprefixerCascade) {
      decl._autoprefixerCascade =
        this.all.options.cascade !== false && decl.raw('before').includes('\n')
    ***REMOVED***
    return decl._autoprefixerCascade
  ***REMOVED***

  /**
   * Return unprefixed version of property
   */
  normalize(prop) {
    return prop
  ***REMOVED***

  /**
   * Return list of prefixed properties to clean old prefixes
   */
  old(prop, prefix) {
    return [this.prefixed(prop, prefix)]
  ***REMOVED***

  /**
   * Check `value`, that it contain other prefixes, rather than `prefix`
   */
  otherPrefixes(value, prefix) {
    for (let other of Browsers.prefixes()) {
      if (other === prefix) {
        continue
      ***REMOVED***
      if (value.includes(other)) {
        return value.replace(/var\([^)]+\)/, '').includes(other)
      ***REMOVED***
    ***REMOVED***
    return false
  ***REMOVED***

  /**
   * Return prefixed version of property
   */
  prefixed(prop, prefix) {
    return prefix + prop
  ***REMOVED***

  /**
   * Add spaces for visual cascade
   */
  process(decl, result) {
    if (!this.needCascade(decl)) {
      super.process(decl, result)
      return
    ***REMOVED***

    let prefixes = super.process(decl, result)

    if (!prefixes || !prefixes.length) {
      return
    ***REMOVED***

    this.restoreBefore(decl)
    decl.raws.before = this.calcBefore(prefixes, decl)
  ***REMOVED***

  /**
   * Remove visual cascade
   */
  restoreBefore(decl) {
    let lines = decl.raw('before').split('\n')
    let min = lines[lines.length - 1]

    this.all.group(decl).up(prefixed => {
      let array = prefixed.raw('before').split('\n')
      let last = array[array.length - 1]
      if (last.length < min.length) {
        min = last
      ***REMOVED***
    ***REMOVED***)

    lines[lines.length - 1] = min
    decl.raws.before = lines.join('\n')
  ***REMOVED***

  /**
   * Set prefix to declaration
   */
  set(decl, prefix) {
    decl.prop = this.prefixed(decl.prop, prefix)
    return decl
  ***REMOVED***
***REMOVED***

module.exports = Declaration
