// autoprefixer/lib/info.js
let browserslist = require('browserslist')

function capitalize(str) {
  return str.slice(0, 1).toUpperCase() + str.slice(1)
***REMOVED***

const NAMES = {
  and_chr: 'Chrome for Android',
  and_ff: 'Firefox for Android',
  and_qq: 'QQ Browser',
  and_uc: 'UC for Android',
  baidu: 'Baidu Browser',
  ie: 'IE',
  ie_mob: 'IE Mobile',
  ios_saf: 'iOS Safari',
  kaios: 'KaiOS Browser',
  op_mini: 'Opera Mini',
  op_mob: 'Opera Mobile',
  samsung: 'Samsung Internet'
***REMOVED***

function prefix(name, prefixes, note) {
  let out = `  ${name***REMOVED***`
  if (note) out += ' *'
  out += ': '
  out += prefixes.map(i => i.replace(/^-(.*)-$/g, '$1')).join(', ')
  out += '\n'
  return out
***REMOVED***

module.exports = function (prefixes) {
  if (prefixes.browsers.selected.length === 0) {
    return 'No browsers selected'
  ***REMOVED***

  let versions = {***REMOVED***
  for (let browser of prefixes.browsers.selected) {
    let parts = browser.split(' ')
    let name = parts[0]
    let version = parts[1]

    name = NAMES[name] || capitalize(name)
    if (versions[name]) {
      versions[name].push(version)
    ***REMOVED*** else {
      versions[name] = [version]
    ***REMOVED***
  ***REMOVED***

  let out = 'Browsers:\n'
  for (let browser in versions) {
    let list = versions[browser]
    list = list.sort((a, b) => parseFloat(b) - parseFloat(a))
    out += `  ${browser***REMOVED***: ${list.join(', ')***REMOVED***\n`
  ***REMOVED***

  let coverage = browserslist.coverage(prefixes.browsers.selected)
  let round = Math.round(coverage * 100) / 100.0
  out += `\nThese browsers account for ${round***REMOVED***% of all users globally\n`

  let atrules = []
  for (let name in prefixes.add) {
    let data = prefixes.add[name]
    if (name[0] === '@' && data.prefixes) {
      atrules.push(prefix(name, data.prefixes))
    ***REMOVED***
  ***REMOVED***
  if (atrules.length > 0) {
    out += `\nAt-Rules:\n${atrules.sort().join('')***REMOVED***`
  ***REMOVED***

  let selectors = []
  for (let selector of prefixes.add.selectors) {
    if (selector.prefixes) {
      selectors.push(prefix(selector.name, selector.prefixes))
    ***REMOVED***
  ***REMOVED***
  if (selectors.length > 0) {
    out += `\nSelectors:\n${selectors.sort().join('')***REMOVED***`
  ***REMOVED***

  let values = []
  let props = []
  let hadGrid = false
  for (let name in prefixes.add) {
    let data = prefixes.add[name]
    if (name[0] !== '@' && data.prefixes) {
      let grid = name.indexOf('grid-') === 0
      if (grid) hadGrid = true
      props.push(prefix(name, data.prefixes, grid))
    ***REMOVED***

    if (!Array.isArray(data.values)) {
      continue
    ***REMOVED***
    for (let value of data.values) {
      let grid = value.name.includes('grid')
      if (grid) hadGrid = true
      let string = prefix(value.name, value.prefixes, grid)
      if (!values.includes(string)) {
        values.push(string)
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  if (props.length > 0) {
    out += `\nProperties:\n${props.sort().join('')***REMOVED***`
  ***REMOVED***
  if (values.length > 0) {
    out += `\nValues:\n${values.sort().join('')***REMOVED***`
  ***REMOVED***
  if (hadGrid) {
    out += '\n* - Prefixes will be added only on grid: true option.\n'
  ***REMOVED***

  if (!atrules.length && !selectors.length && !props.length && !values.length) {
    out +=
      "\nAwesome! Your browsers don't require any vendor prefixes." +
      '\nNow you can remove Autoprefixer from build steps.'
  ***REMOVED***

  return out
***REMOVED***
