// autoprefixer/lib/old-selector.js
class OldSelector {
  constructor(selector, prefix) {
    this.prefix = prefix
    this.prefixed = selector.prefixed(this.prefix)
    this.regexp = selector.regexp(this.prefix)

    this.prefixeds = selector
      .possible()
      .map(x => [selector.prefixed(x), selector.regexp(x)])

    this.unprefixed = selector.name
    this.nameRegexp = selector.regexp()
  ***REMOVED***

  /**
   * Does rule contain an unnecessary prefixed selector
   */
  check(rule) {
    if (!rule.selector.includes(this.prefixed)) {
      return false
    ***REMOVED***
    if (!rule.selector.match(this.regexp)) {
      return false
    ***REMOVED***
    if (this.isHack(rule)) {
      return false
    ***REMOVED***
    return true
  ***REMOVED***

  /**
   * Is rule a hack without unprefixed version bottom
   */
  isHack(rule) {
    let index = rule.parent.index(rule) + 1
    let rules = rule.parent.nodes

    while (index < rules.length) {
      let before = rules[index].selector
      if (!before) {
        return true
      ***REMOVED***

      if (before.includes(this.unprefixed) && before.match(this.nameRegexp)) {
        return false
      ***REMOVED***

      let some = false
      for (let [string, regexp] of this.prefixeds) {
        if (before.includes(string) && before.match(regexp)) {
          some = true
          break
        ***REMOVED***
      ***REMOVED***

      if (!some) {
        return true
      ***REMOVED***

      index += 1
    ***REMOVED***

    return true
  ***REMOVED***
***REMOVED***

module.exports = OldSelector
