// autoprefixer/lib/selector.js
let { list ***REMOVED*** = require('postcss')

let Browsers = require('./browsers')
let OldSelector = require('./old-selector')
let Prefixer = require('./prefixer')
let utils = require('./utils')

class Selector extends Prefixer {
  constructor(name, prefixes, all) {
    super(name, prefixes, all)
    this.regexpCache = new Map()
  ***REMOVED***

  /**
   * Clone and add prefixes for at-rule
   */
  add(rule, prefix) {
    let prefixeds = this.prefixeds(rule)

    if (this.already(rule, prefixeds, prefix)) {
      return
    ***REMOVED***

    let cloned = this.clone(rule, { selector: prefixeds[this.name][prefix] ***REMOVED***)
    rule.parent.insertBefore(rule, cloned)
  ***REMOVED***

  /**
   * Is rule already prefixed before
   */
  already(rule, prefixeds, prefix) {
    let index = rule.parent.index(rule) - 1

    while (index >= 0) {
      let before = rule.parent.nodes[index]

      if (before.type !== 'rule') {
        return false
      ***REMOVED***

      let some = false
      for (let key in prefixeds[this.name]) {
        let prefixed = prefixeds[this.name][key]
        if (before.selector === prefixed) {
          if (prefix === key) {
            return true
          ***REMOVED*** else {
            some = true
            break
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
      if (!some) {
        return false
      ***REMOVED***

      index -= 1
    ***REMOVED***

    return false
  ***REMOVED***

  /**
   * Is rule selectors need to be prefixed
   */
  check(rule) {
    if (rule.selector.includes(this.name)) {
      return !!rule.selector.match(this.regexp())
    ***REMOVED***

    return false
  ***REMOVED***

  /**
   * Return function to fast find prefixed selector
   */
  old(prefix) {
    return new OldSelector(this, prefix)
  ***REMOVED***

  /**
   * All possible prefixes
   */
  possible() {
    return Browsers.prefixes()
  ***REMOVED***

  /**
   * Return prefixed version of selector
   */
  prefixed(prefix) {
    return this.name.replace(/^(\W*)/, `$1${prefix***REMOVED***`)
  ***REMOVED***

  /**
   * Return all possible selector prefixes
   */
  prefixeds(rule) {
    if (rule._autoprefixerPrefixeds) {
      if (rule._autoprefixerPrefixeds[this.name]) {
        return rule._autoprefixerPrefixeds
      ***REMOVED***
    ***REMOVED*** else {
      rule._autoprefixerPrefixeds = {***REMOVED***
    ***REMOVED***

    let prefixeds = {***REMOVED***
    if (rule.selector.includes(',')) {
      let ruleParts = list.comma(rule.selector)
      let toProcess = ruleParts.filter(el => el.includes(this.name))

      for (let prefix of this.possible()) {
        prefixeds[prefix] = toProcess
          .map(el => this.replace(el, prefix))
          .join(', ')
      ***REMOVED***
    ***REMOVED*** else {
      for (let prefix of this.possible()) {
        prefixeds[prefix] = this.replace(rule.selector, prefix)
      ***REMOVED***
    ***REMOVED***

    rule._autoprefixerPrefixeds[this.name] = prefixeds
    return rule._autoprefixerPrefixeds
  ***REMOVED***

  /**
   * Lazy loadRegExp for name
   */
  regexp(prefix) {
    if (!this.regexpCache.has(prefix)) {
      let name = prefix ? this.prefixed(prefix) : this.name
      this.regexpCache.set(
        prefix,
        new RegExp(`(^|[^:"'=])${utils.escapeRegexp(name)***REMOVED***`, 'gi')
      )
    ***REMOVED***

    return this.regexpCache.get(prefix)
  ***REMOVED***

  /**
   * Replace selectors by prefixed one
   */
  replace(selector, prefix) {
    return selector.replace(this.regexp(), `$1${this.prefixed(prefix)***REMOVED***`)
  ***REMOVED***
***REMOVED***

module.exports = Selector
