// jsdom/lib/jsdom/living/generated/AbstractRange.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;

const interfaceName = "AbstractRange";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'AbstractRange'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["AbstractRange"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor AbstractRange is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***
  class AbstractRange {
    constructor() {
      throw new TypeError("Illegal constructor");
    ***REMOVED***

    get startContainer() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get startContainer' called on an object that is not a valid instance of AbstractRange.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["startContainer"]);
    ***REMOVED***

    get startOffset() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get startOffset' called on an object that is not a valid instance of AbstractRange.");
      ***REMOVED***

      return esValue[implSymbol]["startOffset"];
    ***REMOVED***

    get endContainer() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get endContainer' called on an object that is not a valid instance of AbstractRange.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["endContainer"]);
    ***REMOVED***

    get endOffset() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get endOffset' called on an object that is not a valid instance of AbstractRange.");
      ***REMOVED***

      return esValue[implSymbol]["endOffset"];
    ***REMOVED***

    get collapsed() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get collapsed' called on an object that is not a valid instance of AbstractRange.");
      ***REMOVED***

      return esValue[implSymbol]["collapsed"];
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(AbstractRange.prototype, {
    startContainer: { enumerable: true ***REMOVED***,
    startOffset: { enumerable: true ***REMOVED***,
    endContainer: { enumerable: true ***REMOVED***,
    endOffset: { enumerable: true ***REMOVED***,
    collapsed: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "AbstractRange", configurable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = AbstractRange;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: AbstractRange
  ***REMOVED***);
***REMOVED***;

const Impl = require("../range/AbstractRange-impl.js");
