// jsdom/lib/jsdom/living/generated/Document.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const ElementCreationOptions = require("./ElementCreationOptions.js");
const ceReactionsPreSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPreSteps;
const ceReactionsPostSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPostSteps;
const Node = require("./Node.js");
const NodeFilter = require("./NodeFilter.js");
const HTMLElement = require("./HTMLElement.js");
const EventHandlerNonNull = require("./EventHandlerNonNull.js");
const OnErrorEventHandlerNonNull = require("./OnErrorEventHandlerNonNull.js");
const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;

const interfaceName = "Document";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'Document'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["Document"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor Document is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {
  Node._internalSetup(wrapper, globalObject);

  Object.defineProperties(
    wrapper,
    Object.getOwnPropertyDescriptors({
      get location() {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get location' called on an object that is not a valid instance of Document.");
        ***REMOVED***

        return utils.tryWrapperForImpl(esValue[implSymbol]["location"]);
      ***REMOVED***,
      set location(V) {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set location' called on an object that is not a valid instance of Document.");
        ***REMOVED***

        const Q = esValue["location"];
        if (!utils.isObject(Q)) {
          throw new TypeError("Property 'location' is not an object");
        ***REMOVED***
        Reflect.set(Q, "href", V);
      ***REMOVED***
    ***REMOVED***)
  );

  Object.defineProperties(wrapper, { location: { configurable: false ***REMOVED*** ***REMOVED***);
***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***

  if (globalObject.Node === undefined) {
    throw new Error("Internal error: attempting to evaluate Document before Node");
  ***REMOVED***
  class Document extends globalObject.Node {
    constructor() {
      return exports.setup(Object.create(new.target.prototype), globalObject, undefined);
    ***REMOVED***

    getElementsByTagName(qualifiedName) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'getElementsByTagName' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'getElementsByTagName' on 'Document': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'getElementsByTagName' on 'Document': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].getElementsByTagName(...args));
    ***REMOVED***

    getElementsByTagNameNS(namespace, localName) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'getElementsByTagNameNS' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 2) {
        throw new TypeError(
          "Failed to execute 'getElementsByTagNameNS' on 'Document': 2 arguments required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        if (curArg === null || curArg === undefined) {
          curArg = null;
        ***REMOVED*** else {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'getElementsByTagNameNS' on 'Document': parameter 1"
          ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'getElementsByTagNameNS' on 'Document': parameter 2"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].getElementsByTagNameNS(...args));
    ***REMOVED***

    getElementsByClassName(classNames) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'getElementsByClassName' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'getElementsByClassName' on 'Document': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'getElementsByClassName' on 'Document': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].getElementsByClassName(...args));
    ***REMOVED***

    createElement(localName) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'createElement' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'createElement' on 'Document': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'createElement' on 'Document': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        if (curArg !== undefined) {
          if (curArg === null || curArg === undefined) {
            curArg = ElementCreationOptions.convert(curArg, {
              context: "Failed to execute 'createElement' on 'Document': parameter 2"
            ***REMOVED***);
          ***REMOVED*** else if (utils.isObject(curArg)) {
            curArg = ElementCreationOptions.convert(curArg, {
              context: "Failed to execute 'createElement' on 'Document': parameter 2" + " dictionary"
            ***REMOVED***);
          ***REMOVED*** else {
            curArg = conversions["DOMString"](curArg, {
              context: "Failed to execute 'createElement' on 'Document': parameter 2"
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return utils.tryWrapperForImpl(esValue[implSymbol].createElement(...args));
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    createElementNS(namespace, qualifiedName) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'createElementNS' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 2) {
        throw new TypeError(
          "Failed to execute 'createElementNS' on 'Document': 2 arguments required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        if (curArg === null || curArg === undefined) {
          curArg = null;
        ***REMOVED*** else {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'createElementNS' on 'Document': parameter 1"
          ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'createElementNS' on 'Document': parameter 2"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[2];
        if (curArg !== undefined) {
          if (curArg === null || curArg === undefined) {
            curArg = ElementCreationOptions.convert(curArg, {
              context: "Failed to execute 'createElementNS' on 'Document': parameter 3"
            ***REMOVED***);
          ***REMOVED*** else if (utils.isObject(curArg)) {
            curArg = ElementCreationOptions.convert(curArg, {
              context: "Failed to execute 'createElementNS' on 'Document': parameter 3" + " dictionary"
            ***REMOVED***);
          ***REMOVED*** else {
            curArg = conversions["DOMString"](curArg, {
              context: "Failed to execute 'createElementNS' on 'Document': parameter 3"
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return utils.tryWrapperForImpl(esValue[implSymbol].createElementNS(...args));
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    createDocumentFragment() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'createDocumentFragment' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol].createDocumentFragment());
    ***REMOVED***

    createTextNode(data) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'createTextNode' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'createTextNode' on 'Document': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'createTextNode' on 'Document': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].createTextNode(...args));
    ***REMOVED***

    createCDATASection(data) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'createCDATASection' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'createCDATASection' on 'Document': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'createCDATASection' on 'Document': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].createCDATASection(...args));
    ***REMOVED***

    createComment(data) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'createComment' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'createComment' on 'Document': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'createComment' on 'Document': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].createComment(...args));
    ***REMOVED***

    createProcessingInstruction(target, data) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError(
          "'createProcessingInstruction' called on an object that is not a valid instance of Document."
        );
      ***REMOVED***

      if (arguments.length < 2) {
        throw new TypeError(
          "Failed to execute 'createProcessingInstruction' on 'Document': 2 arguments required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'createProcessingInstruction' on 'Document': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'createProcessingInstruction' on 'Document': parameter 2"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].createProcessingInstruction(...args));
    ***REMOVED***

    importNode(node) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'importNode' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'importNode' on 'Document': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = Node.convert(curArg, { context: "Failed to execute 'importNode' on 'Document': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        if (curArg !== undefined) {
          curArg = conversions["boolean"](curArg, {
            context: "Failed to execute 'importNode' on 'Document': parameter 2"
          ***REMOVED***);
        ***REMOVED*** else {
          curArg = false;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return utils.tryWrapperForImpl(esValue[implSymbol].importNode(...args));
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    adoptNode(node) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'adoptNode' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'adoptNode' on 'Document': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = Node.convert(curArg, { context: "Failed to execute 'adoptNode' on 'Document': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return utils.tryWrapperForImpl(esValue[implSymbol].adoptNode(...args));
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    createAttribute(localName) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'createAttribute' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'createAttribute' on 'Document': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'createAttribute' on 'Document': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].createAttribute(...args));
    ***REMOVED***

    createAttributeNS(namespace, qualifiedName) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'createAttributeNS' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 2) {
        throw new TypeError(
          "Failed to execute 'createAttributeNS' on 'Document': 2 arguments required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        if (curArg === null || curArg === undefined) {
          curArg = null;
        ***REMOVED*** else {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'createAttributeNS' on 'Document': parameter 1"
          ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'createAttributeNS' on 'Document': parameter 2"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].createAttributeNS(...args));
    ***REMOVED***

    createEvent(interface_) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'createEvent' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'createEvent' on 'Document': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'createEvent' on 'Document': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].createEvent(...args));
    ***REMOVED***

    createRange() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'createRange' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol].createRange());
    ***REMOVED***

    createNodeIterator(root) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'createNodeIterator' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'createNodeIterator' on 'Document': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = Node.convert(curArg, { context: "Failed to execute 'createNodeIterator' on 'Document': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        if (curArg !== undefined) {
          curArg = conversions["unsigned long"](curArg, {
            context: "Failed to execute 'createNodeIterator' on 'Document': parameter 2"
          ***REMOVED***);
        ***REMOVED*** else {
          curArg = 0xffffffff;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[2];
        if (curArg !== undefined) {
          if (curArg === null || curArg === undefined) {
            curArg = null;
          ***REMOVED*** else {
            curArg = NodeFilter.convert(curArg, {
              context: "Failed to execute 'createNodeIterator' on 'Document': parameter 3"
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED*** else {
          curArg = null;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].createNodeIterator(...args));
    ***REMOVED***

    createTreeWalker(root) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'createTreeWalker' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'createTreeWalker' on 'Document': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = Node.convert(curArg, { context: "Failed to execute 'createTreeWalker' on 'Document': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        if (curArg !== undefined) {
          curArg = conversions["unsigned long"](curArg, {
            context: "Failed to execute 'createTreeWalker' on 'Document': parameter 2"
          ***REMOVED***);
        ***REMOVED*** else {
          curArg = 0xffffffff;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[2];
        if (curArg !== undefined) {
          if (curArg === null || curArg === undefined) {
            curArg = null;
          ***REMOVED*** else {
            curArg = NodeFilter.convert(curArg, {
              context: "Failed to execute 'createTreeWalker' on 'Document': parameter 3"
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED*** else {
          curArg = null;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].createTreeWalker(...args));
    ***REMOVED***

    getElementsByName(elementName) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'getElementsByName' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'getElementsByName' on 'Document': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'getElementsByName' on 'Document': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].getElementsByName(...args));
    ***REMOVED***

    open() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'open' called on an object that is not a valid instance of Document.");
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        if (curArg !== undefined) {
          curArg = conversions["DOMString"](curArg, { context: "Failed to execute 'open' on 'Document': parameter 1" ***REMOVED***);
        ***REMOVED*** else {
          curArg = "text/html";
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        if (curArg !== undefined) {
          curArg = conversions["DOMString"](curArg, { context: "Failed to execute 'open' on 'Document': parameter 2" ***REMOVED***);
        ***REMOVED*** else {
          curArg = "";
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return utils.tryWrapperForImpl(esValue[implSymbol].open(...args));
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    close() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'close' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].close();
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    write() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'write' called on an object that is not a valid instance of Document.");
      ***REMOVED***
      const args = [];
      for (let i = 0; i < arguments.length; i++) {
        let curArg = arguments[i];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'write' on 'Document': parameter " + (i + 1)
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].write(...args);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    writeln() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'writeln' called on an object that is not a valid instance of Document.");
      ***REMOVED***
      const args = [];
      for (let i = 0; i < arguments.length; i++) {
        let curArg = arguments[i];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'writeln' on 'Document': parameter " + (i + 1)
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].writeln(...args);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    hasFocus() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'hasFocus' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return esValue[implSymbol].hasFocus();
    ***REMOVED***

    clear() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'clear' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return esValue[implSymbol].clear();
    ***REMOVED***

    captureEvents() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'captureEvents' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return esValue[implSymbol].captureEvents();
    ***REMOVED***

    releaseEvents() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'releaseEvents' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return esValue[implSymbol].releaseEvents();
    ***REMOVED***

    getSelection() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'getSelection' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol].getSelection());
    ***REMOVED***

    getElementById(elementId) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'getElementById' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'getElementById' on 'Document': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'getElementById' on 'Document': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].getElementById(...args));
    ***REMOVED***

    prepend() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'prepend' called on an object that is not a valid instance of Document.");
      ***REMOVED***
      const args = [];
      for (let i = 0; i < arguments.length; i++) {
        let curArg = arguments[i];
        if (Node.is(curArg)) {
          curArg = utils.implForWrapper(curArg);
        ***REMOVED*** else {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'prepend' on 'Document': parameter " + (i + 1)
          ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].prepend(...args);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    append() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'append' called on an object that is not a valid instance of Document.");
      ***REMOVED***
      const args = [];
      for (let i = 0; i < arguments.length; i++) {
        let curArg = arguments[i];
        if (Node.is(curArg)) {
          curArg = utils.implForWrapper(curArg);
        ***REMOVED*** else {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'append' on 'Document': parameter " + (i + 1)
          ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].append(...args);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    replaceChildren() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'replaceChildren' called on an object that is not a valid instance of Document.");
      ***REMOVED***
      const args = [];
      for (let i = 0; i < arguments.length; i++) {
        let curArg = arguments[i];
        if (Node.is(curArg)) {
          curArg = utils.implForWrapper(curArg);
        ***REMOVED*** else {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'replaceChildren' on 'Document': parameter " + (i + 1)
          ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return utils.tryWrapperForImpl(esValue[implSymbol].replaceChildren(...args));
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    querySelector(selectors) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'querySelector' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'querySelector' on 'Document': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'querySelector' on 'Document': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].querySelector(...args));
    ***REMOVED***

    querySelectorAll(selectors) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'querySelectorAll' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'querySelectorAll' on 'Document': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'querySelectorAll' on 'Document': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].querySelectorAll(...args));
    ***REMOVED***

    get implementation() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get implementation' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.getSameObject(this, "implementation", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["implementation"]);
      ***REMOVED***);
    ***REMOVED***

    get URL() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get URL' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return esValue[implSymbol]["URL"];
    ***REMOVED***

    get documentURI() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get documentURI' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return esValue[implSymbol]["documentURI"];
    ***REMOVED***

    get compatMode() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get compatMode' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return esValue[implSymbol]["compatMode"];
    ***REMOVED***

    get characterSet() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get characterSet' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return esValue[implSymbol]["characterSet"];
    ***REMOVED***

    get charset() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get charset' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return esValue[implSymbol]["charset"];
    ***REMOVED***

    get inputEncoding() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get inputEncoding' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return esValue[implSymbol]["inputEncoding"];
    ***REMOVED***

    get contentType() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get contentType' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return esValue[implSymbol]["contentType"];
    ***REMOVED***

    get doctype() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get doctype' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["doctype"]);
    ***REMOVED***

    get documentElement() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get documentElement' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["documentElement"]);
    ***REMOVED***

    get referrer() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get referrer' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return esValue[implSymbol]["referrer"];
    ***REMOVED***

    get cookie() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get cookie' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return esValue[implSymbol]["cookie"];
    ***REMOVED***

    set cookie(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set cookie' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      V = conversions["USVString"](V, {
        context: "Failed to set the 'cookie' property on 'Document': The provided value"
      ***REMOVED***);

      esValue[implSymbol]["cookie"] = V;
    ***REMOVED***

    get lastModified() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get lastModified' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return esValue[implSymbol]["lastModified"];
    ***REMOVED***

    get readyState() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get readyState' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["readyState"]);
    ***REMOVED***

    get title() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get title' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol]["title"];
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set title(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set title' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'title' property on 'Document': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol]["title"] = V;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get dir() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get dir' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol]["dir"];
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set dir(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set dir' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'dir' property on 'Document': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol]["dir"] = V;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get body() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get body' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return utils.tryWrapperForImpl(esValue[implSymbol]["body"]);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set body(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set body' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (V === null || V === undefined) {
        V = null;
      ***REMOVED*** else {
        V = HTMLElement.convert(V, { context: "Failed to set the 'body' property on 'Document': The provided value" ***REMOVED***);
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol]["body"] = V;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get head() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get head' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["head"]);
    ***REMOVED***

    get images() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get images' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.getSameObject(this, "images", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["images"]);
      ***REMOVED***);
    ***REMOVED***

    get embeds() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get embeds' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.getSameObject(this, "embeds", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["embeds"]);
      ***REMOVED***);
    ***REMOVED***

    get plugins() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get plugins' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.getSameObject(this, "plugins", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["plugins"]);
      ***REMOVED***);
    ***REMOVED***

    get links() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get links' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.getSameObject(this, "links", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["links"]);
      ***REMOVED***);
    ***REMOVED***

    get forms() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get forms' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.getSameObject(this, "forms", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["forms"]);
      ***REMOVED***);
    ***REMOVED***

    get scripts() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get scripts' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.getSameObject(this, "scripts", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["scripts"]);
      ***REMOVED***);
    ***REMOVED***

    get currentScript() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get currentScript' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["currentScript"]);
    ***REMOVED***

    get defaultView() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get defaultView' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["defaultView"]);
    ***REMOVED***

    get onreadystatechange() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        return;
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onreadystatechange"]);
    ***REMOVED***

    set onreadystatechange(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        return;
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onreadystatechange' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onreadystatechange"] = V;
    ***REMOVED***

    get anchors() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get anchors' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.getSameObject(this, "anchors", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["anchors"]);
      ***REMOVED***);
    ***REMOVED***

    get applets() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get applets' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.getSameObject(this, "applets", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["applets"]);
      ***REMOVED***);
    ***REMOVED***

    get styleSheets() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get styleSheets' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.getSameObject(this, "styleSheets", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["styleSheets"]);
      ***REMOVED***);
    ***REMOVED***

    get hidden() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get hidden' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return esValue[implSymbol]["hidden"];
    ***REMOVED***

    get visibilityState() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get visibilityState' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["visibilityState"]);
    ***REMOVED***

    get onvisibilitychange() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onvisibilitychange' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onvisibilitychange"]);
    ***REMOVED***

    set onvisibilitychange(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onvisibilitychange' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onvisibilitychange' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onvisibilitychange"] = V;
    ***REMOVED***

    get onabort() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onabort' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onabort"]);
    ***REMOVED***

    set onabort(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onabort' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onabort' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onabort"] = V;
    ***REMOVED***

    get onauxclick() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onauxclick' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onauxclick"]);
    ***REMOVED***

    set onauxclick(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onauxclick' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onauxclick' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onauxclick"] = V;
    ***REMOVED***

    get onblur() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onblur' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onblur"]);
    ***REMOVED***

    set onblur(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onblur' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onblur' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onblur"] = V;
    ***REMOVED***

    get oncancel() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get oncancel' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["oncancel"]);
    ***REMOVED***

    set oncancel(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set oncancel' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'oncancel' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["oncancel"] = V;
    ***REMOVED***

    get oncanplay() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get oncanplay' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["oncanplay"]);
    ***REMOVED***

    set oncanplay(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set oncanplay' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'oncanplay' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["oncanplay"] = V;
    ***REMOVED***

    get oncanplaythrough() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get oncanplaythrough' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["oncanplaythrough"]);
    ***REMOVED***

    set oncanplaythrough(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set oncanplaythrough' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'oncanplaythrough' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["oncanplaythrough"] = V;
    ***REMOVED***

    get onchange() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onchange' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onchange"]);
    ***REMOVED***

    set onchange(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onchange' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onchange' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onchange"] = V;
    ***REMOVED***

    get onclick() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onclick' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onclick"]);
    ***REMOVED***

    set onclick(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onclick' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onclick' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onclick"] = V;
    ***REMOVED***

    get onclose() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onclose' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onclose"]);
    ***REMOVED***

    set onclose(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onclose' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onclose' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onclose"] = V;
    ***REMOVED***

    get oncontextmenu() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get oncontextmenu' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["oncontextmenu"]);
    ***REMOVED***

    set oncontextmenu(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set oncontextmenu' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'oncontextmenu' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["oncontextmenu"] = V;
    ***REMOVED***

    get oncuechange() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get oncuechange' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["oncuechange"]);
    ***REMOVED***

    set oncuechange(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set oncuechange' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'oncuechange' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["oncuechange"] = V;
    ***REMOVED***

    get ondblclick() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get ondblclick' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["ondblclick"]);
    ***REMOVED***

    set ondblclick(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set ondblclick' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'ondblclick' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["ondblclick"] = V;
    ***REMOVED***

    get ondrag() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get ondrag' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["ondrag"]);
    ***REMOVED***

    set ondrag(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set ondrag' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'ondrag' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["ondrag"] = V;
    ***REMOVED***

    get ondragend() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get ondragend' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["ondragend"]);
    ***REMOVED***

    set ondragend(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set ondragend' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'ondragend' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["ondragend"] = V;
    ***REMOVED***

    get ondragenter() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get ondragenter' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["ondragenter"]);
    ***REMOVED***

    set ondragenter(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set ondragenter' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'ondragenter' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["ondragenter"] = V;
    ***REMOVED***

    get ondragleave() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get ondragleave' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["ondragleave"]);
    ***REMOVED***

    set ondragleave(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set ondragleave' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'ondragleave' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["ondragleave"] = V;
    ***REMOVED***

    get ondragover() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get ondragover' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["ondragover"]);
    ***REMOVED***

    set ondragover(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set ondragover' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'ondragover' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["ondragover"] = V;
    ***REMOVED***

    get ondragstart() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get ondragstart' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["ondragstart"]);
    ***REMOVED***

    set ondragstart(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set ondragstart' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'ondragstart' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["ondragstart"] = V;
    ***REMOVED***

    get ondrop() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get ondrop' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["ondrop"]);
    ***REMOVED***

    set ondrop(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set ondrop' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'ondrop' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["ondrop"] = V;
    ***REMOVED***

    get ondurationchange() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get ondurationchange' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["ondurationchange"]);
    ***REMOVED***

    set ondurationchange(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set ondurationchange' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'ondurationchange' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["ondurationchange"] = V;
    ***REMOVED***

    get onemptied() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onemptied' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onemptied"]);
    ***REMOVED***

    set onemptied(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onemptied' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onemptied' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onemptied"] = V;
    ***REMOVED***

    get onended() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onended' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onended"]);
    ***REMOVED***

    set onended(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onended' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onended' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onended"] = V;
    ***REMOVED***

    get onerror() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onerror' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onerror"]);
    ***REMOVED***

    set onerror(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onerror' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = OnErrorEventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onerror' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onerror"] = V;
    ***REMOVED***

    get onfocus() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onfocus' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onfocus"]);
    ***REMOVED***

    set onfocus(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onfocus' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onfocus' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onfocus"] = V;
    ***REMOVED***

    get oninput() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get oninput' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["oninput"]);
    ***REMOVED***

    set oninput(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set oninput' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'oninput' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["oninput"] = V;
    ***REMOVED***

    get oninvalid() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get oninvalid' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["oninvalid"]);
    ***REMOVED***

    set oninvalid(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set oninvalid' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'oninvalid' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["oninvalid"] = V;
    ***REMOVED***

    get onkeydown() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onkeydown' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onkeydown"]);
    ***REMOVED***

    set onkeydown(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onkeydown' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onkeydown' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onkeydown"] = V;
    ***REMOVED***

    get onkeypress() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onkeypress' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onkeypress"]);
    ***REMOVED***

    set onkeypress(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onkeypress' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onkeypress' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onkeypress"] = V;
    ***REMOVED***

    get onkeyup() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onkeyup' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onkeyup"]);
    ***REMOVED***

    set onkeyup(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onkeyup' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onkeyup' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onkeyup"] = V;
    ***REMOVED***

    get onload() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onload' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onload"]);
    ***REMOVED***

    set onload(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onload' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onload' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onload"] = V;
    ***REMOVED***

    get onloadeddata() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onloadeddata' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onloadeddata"]);
    ***REMOVED***

    set onloadeddata(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onloadeddata' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onloadeddata' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onloadeddata"] = V;
    ***REMOVED***

    get onloadedmetadata() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onloadedmetadata' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onloadedmetadata"]);
    ***REMOVED***

    set onloadedmetadata(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onloadedmetadata' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onloadedmetadata' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onloadedmetadata"] = V;
    ***REMOVED***

    get onloadend() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onloadend' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onloadend"]);
    ***REMOVED***

    set onloadend(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onloadend' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onloadend' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onloadend"] = V;
    ***REMOVED***

    get onloadstart() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onloadstart' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onloadstart"]);
    ***REMOVED***

    set onloadstart(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onloadstart' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onloadstart' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onloadstart"] = V;
    ***REMOVED***

    get onmousedown() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onmousedown' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onmousedown"]);
    ***REMOVED***

    set onmousedown(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onmousedown' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onmousedown' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onmousedown"] = V;
    ***REMOVED***

    get onmouseenter() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        return;
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onmouseenter"]);
    ***REMOVED***

    set onmouseenter(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        return;
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onmouseenter' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onmouseenter"] = V;
    ***REMOVED***

    get onmouseleave() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        return;
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onmouseleave"]);
    ***REMOVED***

    set onmouseleave(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        return;
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onmouseleave' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onmouseleave"] = V;
    ***REMOVED***

    get onmousemove() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onmousemove' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onmousemove"]);
    ***REMOVED***

    set onmousemove(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onmousemove' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onmousemove' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onmousemove"] = V;
    ***REMOVED***

    get onmouseout() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onmouseout' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onmouseout"]);
    ***REMOVED***

    set onmouseout(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onmouseout' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onmouseout' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onmouseout"] = V;
    ***REMOVED***

    get onmouseover() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onmouseover' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onmouseover"]);
    ***REMOVED***

    set onmouseover(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onmouseover' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onmouseover' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onmouseover"] = V;
    ***REMOVED***

    get onmouseup() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onmouseup' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onmouseup"]);
    ***REMOVED***

    set onmouseup(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onmouseup' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onmouseup' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onmouseup"] = V;
    ***REMOVED***

    get onwheel() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onwheel' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onwheel"]);
    ***REMOVED***

    set onwheel(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onwheel' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onwheel' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onwheel"] = V;
    ***REMOVED***

    get onpause() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onpause' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onpause"]);
    ***REMOVED***

    set onpause(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onpause' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onpause' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onpause"] = V;
    ***REMOVED***

    get onplay() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onplay' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onplay"]);
    ***REMOVED***

    set onplay(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onplay' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onplay' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onplay"] = V;
    ***REMOVED***

    get onplaying() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onplaying' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onplaying"]);
    ***REMOVED***

    set onplaying(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onplaying' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onplaying' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onplaying"] = V;
    ***REMOVED***

    get onprogress() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onprogress' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onprogress"]);
    ***REMOVED***

    set onprogress(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onprogress' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onprogress' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onprogress"] = V;
    ***REMOVED***

    get onratechange() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onratechange' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onratechange"]);
    ***REMOVED***

    set onratechange(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onratechange' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onratechange' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onratechange"] = V;
    ***REMOVED***

    get onreset() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onreset' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onreset"]);
    ***REMOVED***

    set onreset(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onreset' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onreset' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onreset"] = V;
    ***REMOVED***

    get onresize() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onresize' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onresize"]);
    ***REMOVED***

    set onresize(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onresize' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onresize' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onresize"] = V;
    ***REMOVED***

    get onscroll() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onscroll' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onscroll"]);
    ***REMOVED***

    set onscroll(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onscroll' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onscroll' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onscroll"] = V;
    ***REMOVED***

    get onsecuritypolicyviolation() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onsecuritypolicyviolation' called on an object that is not a valid instance of Document."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onsecuritypolicyviolation"]);
    ***REMOVED***

    set onsecuritypolicyviolation(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onsecuritypolicyviolation' called on an object that is not a valid instance of Document."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onsecuritypolicyviolation' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onsecuritypolicyviolation"] = V;
    ***REMOVED***

    get onseeked() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onseeked' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onseeked"]);
    ***REMOVED***

    set onseeked(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onseeked' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onseeked' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onseeked"] = V;
    ***REMOVED***

    get onseeking() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onseeking' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onseeking"]);
    ***REMOVED***

    set onseeking(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onseeking' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onseeking' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onseeking"] = V;
    ***REMOVED***

    get onselect() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onselect' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onselect"]);
    ***REMOVED***

    set onselect(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onselect' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onselect' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onselect"] = V;
    ***REMOVED***

    get onstalled() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onstalled' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onstalled"]);
    ***REMOVED***

    set onstalled(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onstalled' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onstalled' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onstalled"] = V;
    ***REMOVED***

    get onsubmit() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onsubmit' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onsubmit"]);
    ***REMOVED***

    set onsubmit(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onsubmit' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onsubmit' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onsubmit"] = V;
    ***REMOVED***

    get onsuspend() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onsuspend' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onsuspend"]);
    ***REMOVED***

    set onsuspend(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onsuspend' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onsuspend' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onsuspend"] = V;
    ***REMOVED***

    get ontimeupdate() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get ontimeupdate' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["ontimeupdate"]);
    ***REMOVED***

    set ontimeupdate(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set ontimeupdate' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'ontimeupdate' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["ontimeupdate"] = V;
    ***REMOVED***

    get ontoggle() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get ontoggle' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["ontoggle"]);
    ***REMOVED***

    set ontoggle(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set ontoggle' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'ontoggle' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["ontoggle"] = V;
    ***REMOVED***

    get onvolumechange() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onvolumechange' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onvolumechange"]);
    ***REMOVED***

    set onvolumechange(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onvolumechange' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onvolumechange' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onvolumechange"] = V;
    ***REMOVED***

    get onwaiting() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onwaiting' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onwaiting"]);
    ***REMOVED***

    set onwaiting(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onwaiting' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onwaiting' property on 'Document': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onwaiting"] = V;
    ***REMOVED***

    get activeElement() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get activeElement' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["activeElement"]);
    ***REMOVED***

    get children() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get children' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.getSameObject(this, "children", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["children"]);
      ***REMOVED***);
    ***REMOVED***

    get firstElementChild() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get firstElementChild' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["firstElementChild"]);
    ***REMOVED***

    get lastElementChild() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get lastElementChild' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["lastElementChild"]);
    ***REMOVED***

    get childElementCount() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get childElementCount' called on an object that is not a valid instance of Document.");
      ***REMOVED***

      return esValue[implSymbol]["childElementCount"];
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(Document.prototype, {
    getElementsByTagName: { enumerable: true ***REMOVED***,
    getElementsByTagNameNS: { enumerable: true ***REMOVED***,
    getElementsByClassName: { enumerable: true ***REMOVED***,
    createElement: { enumerable: true ***REMOVED***,
    createElementNS: { enumerable: true ***REMOVED***,
    createDocumentFragment: { enumerable: true ***REMOVED***,
    createTextNode: { enumerable: true ***REMOVED***,
    createCDATASection: { enumerable: true ***REMOVED***,
    createComment: { enumerable: true ***REMOVED***,
    createProcessingInstruction: { enumerable: true ***REMOVED***,
    importNode: { enumerable: true ***REMOVED***,
    adoptNode: { enumerable: true ***REMOVED***,
    createAttribute: { enumerable: true ***REMOVED***,
    createAttributeNS: { enumerable: true ***REMOVED***,
    createEvent: { enumerable: true ***REMOVED***,
    createRange: { enumerable: true ***REMOVED***,
    createNodeIterator: { enumerable: true ***REMOVED***,
    createTreeWalker: { enumerable: true ***REMOVED***,
    getElementsByName: { enumerable: true ***REMOVED***,
    open: { enumerable: true ***REMOVED***,
    close: { enumerable: true ***REMOVED***,
    write: { enumerable: true ***REMOVED***,
    writeln: { enumerable: true ***REMOVED***,
    hasFocus: { enumerable: true ***REMOVED***,
    clear: { enumerable: true ***REMOVED***,
    captureEvents: { enumerable: true ***REMOVED***,
    releaseEvents: { enumerable: true ***REMOVED***,
    getSelection: { enumerable: true ***REMOVED***,
    getElementById: { enumerable: true ***REMOVED***,
    prepend: { enumerable: true ***REMOVED***,
    append: { enumerable: true ***REMOVED***,
    replaceChildren: { enumerable: true ***REMOVED***,
    querySelector: { enumerable: true ***REMOVED***,
    querySelectorAll: { enumerable: true ***REMOVED***,
    implementation: { enumerable: true ***REMOVED***,
    URL: { enumerable: true ***REMOVED***,
    documentURI: { enumerable: true ***REMOVED***,
    compatMode: { enumerable: true ***REMOVED***,
    characterSet: { enumerable: true ***REMOVED***,
    charset: { enumerable: true ***REMOVED***,
    inputEncoding: { enumerable: true ***REMOVED***,
    contentType: { enumerable: true ***REMOVED***,
    doctype: { enumerable: true ***REMOVED***,
    documentElement: { enumerable: true ***REMOVED***,
    referrer: { enumerable: true ***REMOVED***,
    cookie: { enumerable: true ***REMOVED***,
    lastModified: { enumerable: true ***REMOVED***,
    readyState: { enumerable: true ***REMOVED***,
    title: { enumerable: true ***REMOVED***,
    dir: { enumerable: true ***REMOVED***,
    body: { enumerable: true ***REMOVED***,
    head: { enumerable: true ***REMOVED***,
    images: { enumerable: true ***REMOVED***,
    embeds: { enumerable: true ***REMOVED***,
    plugins: { enumerable: true ***REMOVED***,
    links: { enumerable: true ***REMOVED***,
    forms: { enumerable: true ***REMOVED***,
    scripts: { enumerable: true ***REMOVED***,
    currentScript: { enumerable: true ***REMOVED***,
    defaultView: { enumerable: true ***REMOVED***,
    onreadystatechange: { enumerable: true ***REMOVED***,
    anchors: { enumerable: true ***REMOVED***,
    applets: { enumerable: true ***REMOVED***,
    styleSheets: { enumerable: true ***REMOVED***,
    hidden: { enumerable: true ***REMOVED***,
    visibilityState: { enumerable: true ***REMOVED***,
    onvisibilitychange: { enumerable: true ***REMOVED***,
    onabort: { enumerable: true ***REMOVED***,
    onauxclick: { enumerable: true ***REMOVED***,
    onblur: { enumerable: true ***REMOVED***,
    oncancel: { enumerable: true ***REMOVED***,
    oncanplay: { enumerable: true ***REMOVED***,
    oncanplaythrough: { enumerable: true ***REMOVED***,
    onchange: { enumerable: true ***REMOVED***,
    onclick: { enumerable: true ***REMOVED***,
    onclose: { enumerable: true ***REMOVED***,
    oncontextmenu: { enumerable: true ***REMOVED***,
    oncuechange: { enumerable: true ***REMOVED***,
    ondblclick: { enumerable: true ***REMOVED***,
    ondrag: { enumerable: true ***REMOVED***,
    ondragend: { enumerable: true ***REMOVED***,
    ondragenter: { enumerable: true ***REMOVED***,
    ondragleave: { enumerable: true ***REMOVED***,
    ondragover: { enumerable: true ***REMOVED***,
    ondragstart: { enumerable: true ***REMOVED***,
    ondrop: { enumerable: true ***REMOVED***,
    ondurationchange: { enumerable: true ***REMOVED***,
    onemptied: { enumerable: true ***REMOVED***,
    onended: { enumerable: true ***REMOVED***,
    onerror: { enumerable: true ***REMOVED***,
    onfocus: { enumerable: true ***REMOVED***,
    oninput: { enumerable: true ***REMOVED***,
    oninvalid: { enumerable: true ***REMOVED***,
    onkeydown: { enumerable: true ***REMOVED***,
    onkeypress: { enumerable: true ***REMOVED***,
    onkeyup: { enumerable: true ***REMOVED***,
    onload: { enumerable: true ***REMOVED***,
    onloadeddata: { enumerable: true ***REMOVED***,
    onloadedmetadata: { enumerable: true ***REMOVED***,
    onloadend: { enumerable: true ***REMOVED***,
    onloadstart: { enumerable: true ***REMOVED***,
    onmousedown: { enumerable: true ***REMOVED***,
    onmouseenter: { enumerable: true ***REMOVED***,
    onmouseleave: { enumerable: true ***REMOVED***,
    onmousemove: { enumerable: true ***REMOVED***,
    onmouseout: { enumerable: true ***REMOVED***,
    onmouseover: { enumerable: true ***REMOVED***,
    onmouseup: { enumerable: true ***REMOVED***,
    onwheel: { enumerable: true ***REMOVED***,
    onpause: { enumerable: true ***REMOVED***,
    onplay: { enumerable: true ***REMOVED***,
    onplaying: { enumerable: true ***REMOVED***,
    onprogress: { enumerable: true ***REMOVED***,
    onratechange: { enumerable: true ***REMOVED***,
    onreset: { enumerable: true ***REMOVED***,
    onresize: { enumerable: true ***REMOVED***,
    onscroll: { enumerable: true ***REMOVED***,
    onsecuritypolicyviolation: { enumerable: true ***REMOVED***,
    onseeked: { enumerable: true ***REMOVED***,
    onseeking: { enumerable: true ***REMOVED***,
    onselect: { enumerable: true ***REMOVED***,
    onstalled: { enumerable: true ***REMOVED***,
    onsubmit: { enumerable: true ***REMOVED***,
    onsuspend: { enumerable: true ***REMOVED***,
    ontimeupdate: { enumerable: true ***REMOVED***,
    ontoggle: { enumerable: true ***REMOVED***,
    onvolumechange: { enumerable: true ***REMOVED***,
    onwaiting: { enumerable: true ***REMOVED***,
    activeElement: { enumerable: true ***REMOVED***,
    children: { enumerable: true ***REMOVED***,
    firstElementChild: { enumerable: true ***REMOVED***,
    lastElementChild: { enumerable: true ***REMOVED***,
    childElementCount: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "Document", configurable: true ***REMOVED***,
    [Symbol.unscopables]: {
      value: { prepend: true, append: true, replaceChildren: true, __proto__: null ***REMOVED***,
      configurable: true
    ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = Document;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: Document
  ***REMOVED***);
***REMOVED***;

const Impl = require("../nodes/Document-impl.js");
