// jsdom/lib/jsdom/living/generated/DocumentFragment.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const Node = require("./Node.js");
const ceReactionsPreSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPreSteps;
const ceReactionsPostSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPostSteps;
const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;

const interfaceName = "DocumentFragment";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'DocumentFragment'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["DocumentFragment"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor DocumentFragment is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {
  Node._internalSetup(wrapper, globalObject);
***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***

  if (globalObject.Node === undefined) {
    throw new Error("Internal error: attempting to evaluate DocumentFragment before Node");
  ***REMOVED***
  class DocumentFragment extends globalObject.Node {
    constructor() {
      return exports.setup(Object.create(new.target.prototype), globalObject, undefined);
    ***REMOVED***

    getElementById(elementId) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'getElementById' called on an object that is not a valid instance of DocumentFragment.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'getElementById' on 'DocumentFragment': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'getElementById' on 'DocumentFragment': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].getElementById(...args));
    ***REMOVED***

    prepend() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'prepend' called on an object that is not a valid instance of DocumentFragment.");
      ***REMOVED***
      const args = [];
      for (let i = 0; i < arguments.length; i++) {
        let curArg = arguments[i];
        if (Node.is(curArg)) {
          curArg = utils.implForWrapper(curArg);
        ***REMOVED*** else {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'prepend' on 'DocumentFragment': parameter " + (i + 1)
          ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].prepend(...args);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    append() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'append' called on an object that is not a valid instance of DocumentFragment.");
      ***REMOVED***
      const args = [];
      for (let i = 0; i < arguments.length; i++) {
        let curArg = arguments[i];
        if (Node.is(curArg)) {
          curArg = utils.implForWrapper(curArg);
        ***REMOVED*** else {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'append' on 'DocumentFragment': parameter " + (i + 1)
          ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].append(...args);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    replaceChildren() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'replaceChildren' called on an object that is not a valid instance of DocumentFragment.");
      ***REMOVED***
      const args = [];
      for (let i = 0; i < arguments.length; i++) {
        let curArg = arguments[i];
        if (Node.is(curArg)) {
          curArg = utils.implForWrapper(curArg);
        ***REMOVED*** else {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'replaceChildren' on 'DocumentFragment': parameter " + (i + 1)
          ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return utils.tryWrapperForImpl(esValue[implSymbol].replaceChildren(...args));
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    querySelector(selectors) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'querySelector' called on an object that is not a valid instance of DocumentFragment.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'querySelector' on 'DocumentFragment': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'querySelector' on 'DocumentFragment': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].querySelector(...args));
    ***REMOVED***

    querySelectorAll(selectors) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'querySelectorAll' called on an object that is not a valid instance of DocumentFragment.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'querySelectorAll' on 'DocumentFragment': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'querySelectorAll' on 'DocumentFragment': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].querySelectorAll(...args));
    ***REMOVED***

    get children() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get children' called on an object that is not a valid instance of DocumentFragment.");
      ***REMOVED***

      return utils.getSameObject(this, "children", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["children"]);
      ***REMOVED***);
    ***REMOVED***

    get firstElementChild() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get firstElementChild' called on an object that is not a valid instance of DocumentFragment."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["firstElementChild"]);
    ***REMOVED***

    get lastElementChild() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get lastElementChild' called on an object that is not a valid instance of DocumentFragment."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["lastElementChild"]);
    ***REMOVED***

    get childElementCount() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get childElementCount' called on an object that is not a valid instance of DocumentFragment."
        );
      ***REMOVED***

      return esValue[implSymbol]["childElementCount"];
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(DocumentFragment.prototype, {
    getElementById: { enumerable: true ***REMOVED***,
    prepend: { enumerable: true ***REMOVED***,
    append: { enumerable: true ***REMOVED***,
    replaceChildren: { enumerable: true ***REMOVED***,
    querySelector: { enumerable: true ***REMOVED***,
    querySelectorAll: { enumerable: true ***REMOVED***,
    children: { enumerable: true ***REMOVED***,
    firstElementChild: { enumerable: true ***REMOVED***,
    lastElementChild: { enumerable: true ***REMOVED***,
    childElementCount: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "DocumentFragment", configurable: true ***REMOVED***,
    [Symbol.unscopables]: {
      value: { prepend: true, append: true, replaceChildren: true, __proto__: null ***REMOVED***,
      configurable: true
    ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = DocumentFragment;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: DocumentFragment
  ***REMOVED***);
***REMOVED***;

const Impl = require("../nodes/DocumentFragment-impl.js");
