// jsdom/lib/jsdom/living/generated/Event.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const EventInit = require("./EventInit.js");
const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;

const interfaceName = "Event";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'Event'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["Event"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor Event is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {
  Object.defineProperties(
    wrapper,
    Object.getOwnPropertyDescriptors({
      get isTrusted() {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get isTrusted' called on an object that is not a valid instance of Event.");
        ***REMOVED***

        return esValue[implSymbol]["isTrusted"];
      ***REMOVED***
    ***REMOVED***)
  );

  Object.defineProperties(wrapper, { isTrusted: { configurable: false ***REMOVED*** ***REMOVED***);
***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window", "Worker", "AudioWorklet"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***
  class Event {
    constructor(type) {
      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to construct 'Event': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, { context: "Failed to construct 'Event': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        curArg = EventInit.convert(curArg, { context: "Failed to construct 'Event': parameter 2" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return exports.setup(Object.create(new.target.prototype), globalObject, args);
    ***REMOVED***

    composedPath() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'composedPath' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol].composedPath());
    ***REMOVED***

    stopPropagation() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'stopPropagation' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      return esValue[implSymbol].stopPropagation();
    ***REMOVED***

    stopImmediatePropagation() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'stopImmediatePropagation' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      return esValue[implSymbol].stopImmediatePropagation();
    ***REMOVED***

    preventDefault() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'preventDefault' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      return esValue[implSymbol].preventDefault();
    ***REMOVED***

    initEvent(type) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'initEvent' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'initEvent' on 'Event': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, { context: "Failed to execute 'initEvent' on 'Event': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        if (curArg !== undefined) {
          curArg = conversions["boolean"](curArg, { context: "Failed to execute 'initEvent' on 'Event': parameter 2" ***REMOVED***);
        ***REMOVED*** else {
          curArg = false;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[2];
        if (curArg !== undefined) {
          curArg = conversions["boolean"](curArg, { context: "Failed to execute 'initEvent' on 'Event': parameter 3" ***REMOVED***);
        ***REMOVED*** else {
          curArg = false;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].initEvent(...args);
    ***REMOVED***

    get type() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get type' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      return esValue[implSymbol]["type"];
    ***REMOVED***

    get target() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get target' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["target"]);
    ***REMOVED***

    get srcElement() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get srcElement' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["srcElement"]);
    ***REMOVED***

    get currentTarget() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get currentTarget' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["currentTarget"]);
    ***REMOVED***

    get eventPhase() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get eventPhase' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      return esValue[implSymbol]["eventPhase"];
    ***REMOVED***

    get cancelBubble() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get cancelBubble' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      return esValue[implSymbol]["cancelBubble"];
    ***REMOVED***

    set cancelBubble(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set cancelBubble' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'cancelBubble' property on 'Event': The provided value"
      ***REMOVED***);

      esValue[implSymbol]["cancelBubble"] = V;
    ***REMOVED***

    get bubbles() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get bubbles' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      return esValue[implSymbol]["bubbles"];
    ***REMOVED***

    get cancelable() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get cancelable' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      return esValue[implSymbol]["cancelable"];
    ***REMOVED***

    get returnValue() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get returnValue' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      return esValue[implSymbol]["returnValue"];
    ***REMOVED***

    set returnValue(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set returnValue' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'returnValue' property on 'Event': The provided value"
      ***REMOVED***);

      esValue[implSymbol]["returnValue"] = V;
    ***REMOVED***

    get defaultPrevented() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get defaultPrevented' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      return esValue[implSymbol]["defaultPrevented"];
    ***REMOVED***

    get composed() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get composed' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      return esValue[implSymbol]["composed"];
    ***REMOVED***

    get timeStamp() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get timeStamp' called on an object that is not a valid instance of Event.");
      ***REMOVED***

      return esValue[implSymbol]["timeStamp"];
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(Event.prototype, {
    composedPath: { enumerable: true ***REMOVED***,
    stopPropagation: { enumerable: true ***REMOVED***,
    stopImmediatePropagation: { enumerable: true ***REMOVED***,
    preventDefault: { enumerable: true ***REMOVED***,
    initEvent: { enumerable: true ***REMOVED***,
    type: { enumerable: true ***REMOVED***,
    target: { enumerable: true ***REMOVED***,
    srcElement: { enumerable: true ***REMOVED***,
    currentTarget: { enumerable: true ***REMOVED***,
    eventPhase: { enumerable: true ***REMOVED***,
    cancelBubble: { enumerable: true ***REMOVED***,
    bubbles: { enumerable: true ***REMOVED***,
    cancelable: { enumerable: true ***REMOVED***,
    returnValue: { enumerable: true ***REMOVED***,
    defaultPrevented: { enumerable: true ***REMOVED***,
    composed: { enumerable: true ***REMOVED***,
    timeStamp: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "Event", configurable: true ***REMOVED***,
    NONE: { value: 0, enumerable: true ***REMOVED***,
    CAPTURING_PHASE: { value: 1, enumerable: true ***REMOVED***,
    AT_TARGET: { value: 2, enumerable: true ***REMOVED***,
    BUBBLING_PHASE: { value: 3, enumerable: true ***REMOVED***
  ***REMOVED***);
  Object.defineProperties(Event, {
    NONE: { value: 0, enumerable: true ***REMOVED***,
    CAPTURING_PHASE: { value: 1, enumerable: true ***REMOVED***,
    AT_TARGET: { value: 2, enumerable: true ***REMOVED***,
    BUBBLING_PHASE: { value: 3, enumerable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = Event;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: Event
  ***REMOVED***);
***REMOVED***;

const Impl = require("../events/Event-impl.js");
