// jsdom/lib/jsdom/living/generated/FileReader.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const Blob = require("./Blob.js");
const EventHandlerNonNull = require("./EventHandlerNonNull.js");
const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;
const EventTarget = require("./EventTarget.js");

const interfaceName = "FileReader";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'FileReader'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["FileReader"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor FileReader is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {
  EventTarget._internalSetup(wrapper, globalObject);
***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window", "Worker"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***

  if (globalObject.EventTarget === undefined) {
    throw new Error("Internal error: attempting to evaluate FileReader before EventTarget");
  ***REMOVED***
  class FileReader extends globalObject.EventTarget {
    constructor() {
      return exports.setup(Object.create(new.target.prototype), globalObject, undefined);
    ***REMOVED***

    readAsArrayBuffer(blob) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'readAsArrayBuffer' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'readAsArrayBuffer' on 'FileReader': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = Blob.convert(curArg, {
          context: "Failed to execute 'readAsArrayBuffer' on 'FileReader': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].readAsArrayBuffer(...args);
    ***REMOVED***

    readAsBinaryString(blob) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'readAsBinaryString' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'readAsBinaryString' on 'FileReader': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = Blob.convert(curArg, {
          context: "Failed to execute 'readAsBinaryString' on 'FileReader': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].readAsBinaryString(...args);
    ***REMOVED***

    readAsText(blob) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'readAsText' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'readAsText' on 'FileReader': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = Blob.convert(curArg, { context: "Failed to execute 'readAsText' on 'FileReader': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        if (curArg !== undefined) {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'readAsText' on 'FileReader': parameter 2"
          ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].readAsText(...args);
    ***REMOVED***

    readAsDataURL(blob) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'readAsDataURL' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'readAsDataURL' on 'FileReader': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = Blob.convert(curArg, { context: "Failed to execute 'readAsDataURL' on 'FileReader': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].readAsDataURL(...args);
    ***REMOVED***

    abort() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'abort' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      return esValue[implSymbol].abort();
    ***REMOVED***

    get readyState() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get readyState' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      return esValue[implSymbol]["readyState"];
    ***REMOVED***

    get result() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get result' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["result"]);
    ***REMOVED***

    get error() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get error' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["error"]);
    ***REMOVED***

    get onloadstart() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onloadstart' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onloadstart"]);
    ***REMOVED***

    set onloadstart(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onloadstart' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onloadstart' property on 'FileReader': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onloadstart"] = V;
    ***REMOVED***

    get onprogress() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onprogress' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onprogress"]);
    ***REMOVED***

    set onprogress(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onprogress' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onprogress' property on 'FileReader': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onprogress"] = V;
    ***REMOVED***

    get onload() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onload' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onload"]);
    ***REMOVED***

    set onload(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onload' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onload' property on 'FileReader': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onload"] = V;
    ***REMOVED***

    get onabort() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onabort' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onabort"]);
    ***REMOVED***

    set onabort(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onabort' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onabort' property on 'FileReader': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onabort"] = V;
    ***REMOVED***

    get onerror() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onerror' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onerror"]);
    ***REMOVED***

    set onerror(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onerror' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onerror' property on 'FileReader': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onerror"] = V;
    ***REMOVED***

    get onloadend() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onloadend' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onloadend"]);
    ***REMOVED***

    set onloadend(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onloadend' called on an object that is not a valid instance of FileReader.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onloadend' property on 'FileReader': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onloadend"] = V;
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(FileReader.prototype, {
    readAsArrayBuffer: { enumerable: true ***REMOVED***,
    readAsBinaryString: { enumerable: true ***REMOVED***,
    readAsText: { enumerable: true ***REMOVED***,
    readAsDataURL: { enumerable: true ***REMOVED***,
    abort: { enumerable: true ***REMOVED***,
    readyState: { enumerable: true ***REMOVED***,
    result: { enumerable: true ***REMOVED***,
    error: { enumerable: true ***REMOVED***,
    onloadstart: { enumerable: true ***REMOVED***,
    onprogress: { enumerable: true ***REMOVED***,
    onload: { enumerable: true ***REMOVED***,
    onabort: { enumerable: true ***REMOVED***,
    onerror: { enumerable: true ***REMOVED***,
    onloadend: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "FileReader", configurable: true ***REMOVED***,
    EMPTY: { value: 0, enumerable: true ***REMOVED***,
    LOADING: { value: 1, enumerable: true ***REMOVED***,
    DONE: { value: 2, enumerable: true ***REMOVED***
  ***REMOVED***);
  Object.defineProperties(FileReader, {
    EMPTY: { value: 0, enumerable: true ***REMOVED***,
    LOADING: { value: 1, enumerable: true ***REMOVED***,
    DONE: { value: 2, enumerable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = FileReader;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: FileReader
  ***REMOVED***);
***REMOVED***;

const Impl = require("../file-api/FileReader-impl.js");
