// jsdom/lib/jsdom/living/generated/FormData.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const HTMLFormElement = require("./HTMLFormElement.js");
const Blob = require("./Blob.js");
const Function = require("./Function.js");
const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;

const interfaceName = "FormData";

const IteratorPrototype = Object.create(utils.IteratorPrototype, {
  next: {
    value: function next() {
      const internal = this && this[utils.iterInternalSymbol];
      if (!internal) {
        throw new TypeError("next() called on a value that is not an iterator prototype object");
      ***REMOVED***

      const { target, kind, index ***REMOVED*** = internal;
      const values = Array.from(target[implSymbol]);
      const len = values.length;
      if (index >= len) {
        return { value: undefined, done: true ***REMOVED***;
      ***REMOVED***

      const pair = values[index];
      internal.index = index + 1;
      return utils.iteratorResult(pair.map(utils.tryWrapperForImpl), kind);
    ***REMOVED***,
    writable: true,
    enumerable: true,
    configurable: true
  ***REMOVED***,
  [Symbol.toStringTag]: {
    value: "FormData Iterator",
    configurable: true
  ***REMOVED***
***REMOVED***);

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'FormData'.`);
***REMOVED***;

exports.createDefaultIterator = (target, kind) => {
  const iterator = Object.create(IteratorPrototype);
  Object.defineProperty(iterator, utils.iterInternalSymbol, {
    value: { target, kind, index: 0 ***REMOVED***,
    configurable: true
  ***REMOVED***);
  return iterator;
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["FormData"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor FormData is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window", "Worker"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***
  class FormData {
    constructor() {
      const args = [];
      {
        let curArg = arguments[0];
        if (curArg !== undefined) {
          curArg = HTMLFormElement.convert(curArg, { context: "Failed to construct 'FormData': parameter 1" ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return exports.setup(Object.create(new.target.prototype), globalObject, args);
    ***REMOVED***

    append(name, value) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'append' called on an object that is not a valid instance of FormData.");
      ***REMOVED***

      if (arguments.length < 2) {
        throw new TypeError(
          "Failed to execute 'append' on 'FormData': 2 arguments required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      switch (arguments.length) {
        case 2:
          {
            let curArg = arguments[0];
            curArg = conversions["USVString"](curArg, {
              context: "Failed to execute 'append' on 'FormData': parameter 1"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[1];
            if (Blob.is(curArg)) {
              {
                let curArg = arguments[1];
                curArg = Blob.convert(curArg, { context: "Failed to execute 'append' on 'FormData': parameter 2" ***REMOVED***);
                args.push(curArg);
              ***REMOVED***
            ***REMOVED*** else {
              {
                let curArg = arguments[1];
                curArg = conversions["USVString"](curArg, {
                  context: "Failed to execute 'append' on 'FormData': parameter 2"
                ***REMOVED***);
                args.push(curArg);
              ***REMOVED***
            ***REMOVED***
          ***REMOVED***
          break;
        default:
          {
            let curArg = arguments[0];
            curArg = conversions["USVString"](curArg, {
              context: "Failed to execute 'append' on 'FormData': parameter 1"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[1];
            curArg = Blob.convert(curArg, { context: "Failed to execute 'append' on 'FormData': parameter 2" ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[2];
            if (curArg !== undefined) {
              curArg = conversions["USVString"](curArg, {
                context: "Failed to execute 'append' on 'FormData': parameter 3"
              ***REMOVED***);
            ***REMOVED***
            args.push(curArg);
          ***REMOVED***
      ***REMOVED***
      return esValue[implSymbol].append(...args);
    ***REMOVED***

    delete(name) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'delete' called on an object that is not a valid instance of FormData.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'delete' on 'FormData': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["USVString"](curArg, { context: "Failed to execute 'delete' on 'FormData': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].delete(...args);
    ***REMOVED***

    get(name) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'get' called on an object that is not a valid instance of FormData.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'get' on 'FormData': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["USVString"](curArg, { context: "Failed to execute 'get' on 'FormData': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].get(...args));
    ***REMOVED***

    getAll(name) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'getAll' called on an object that is not a valid instance of FormData.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'getAll' on 'FormData': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["USVString"](curArg, { context: "Failed to execute 'getAll' on 'FormData': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].getAll(...args));
    ***REMOVED***

    has(name) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'has' called on an object that is not a valid instance of FormData.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'has' on 'FormData': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["USVString"](curArg, { context: "Failed to execute 'has' on 'FormData': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].has(...args);
    ***REMOVED***

    set(name, value) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'set' called on an object that is not a valid instance of FormData.");
      ***REMOVED***

      if (arguments.length < 2) {
        throw new TypeError(
          "Failed to execute 'set' on 'FormData': 2 arguments required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      switch (arguments.length) {
        case 2:
          {
            let curArg = arguments[0];
            curArg = conversions["USVString"](curArg, {
              context: "Failed to execute 'set' on 'FormData': parameter 1"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[1];
            if (Blob.is(curArg)) {
              {
                let curArg = arguments[1];
                curArg = Blob.convert(curArg, { context: "Failed to execute 'set' on 'FormData': parameter 2" ***REMOVED***);
                args.push(curArg);
              ***REMOVED***
            ***REMOVED*** else {
              {
                let curArg = arguments[1];
                curArg = conversions["USVString"](curArg, {
                  context: "Failed to execute 'set' on 'FormData': parameter 2"
                ***REMOVED***);
                args.push(curArg);
              ***REMOVED***
            ***REMOVED***
          ***REMOVED***
          break;
        default:
          {
            let curArg = arguments[0];
            curArg = conversions["USVString"](curArg, {
              context: "Failed to execute 'set' on 'FormData': parameter 1"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[1];
            curArg = Blob.convert(curArg, { context: "Failed to execute 'set' on 'FormData': parameter 2" ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[2];
            if (curArg !== undefined) {
              curArg = conversions["USVString"](curArg, {
                context: "Failed to execute 'set' on 'FormData': parameter 3"
              ***REMOVED***);
            ***REMOVED***
            args.push(curArg);
          ***REMOVED***
      ***REMOVED***
      return esValue[implSymbol].set(...args);
    ***REMOVED***

    keys() {
      if (!exports.is(this)) {
        throw new TypeError("'keys' called on an object that is not a valid instance of FormData.");
      ***REMOVED***
      return exports.createDefaultIterator(this, "key");
    ***REMOVED***

    values() {
      if (!exports.is(this)) {
        throw new TypeError("'values' called on an object that is not a valid instance of FormData.");
      ***REMOVED***
      return exports.createDefaultIterator(this, "value");
    ***REMOVED***

    entries() {
      if (!exports.is(this)) {
        throw new TypeError("'entries' called on an object that is not a valid instance of FormData.");
      ***REMOVED***
      return exports.createDefaultIterator(this, "key+value");
    ***REMOVED***

    forEach(callback) {
      if (!exports.is(this)) {
        throw new TypeError("'forEach' called on an object that is not a valid instance of FormData.");
      ***REMOVED***
      if (arguments.length < 1) {
        throw new TypeError("Failed to execute 'forEach' on 'iterable': 1 argument required, " + "but only 0 present.");
      ***REMOVED***
      callback = Function.convert(callback, {
        context: "Failed to execute 'forEach' on 'iterable': The callback provided as parameter 1"
      ***REMOVED***);
      const thisArg = arguments[1];
      let pairs = Array.from(this[implSymbol]);
      let i = 0;
      while (i < pairs.length) {
        const [key, value] = pairs[i].map(utils.tryWrapperForImpl);
        callback.call(thisArg, value, key, this);
        pairs = Array.from(this[implSymbol]);
        i++;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(FormData.prototype, {
    append: { enumerable: true ***REMOVED***,
    delete: { enumerable: true ***REMOVED***,
    get: { enumerable: true ***REMOVED***,
    getAll: { enumerable: true ***REMOVED***,
    has: { enumerable: true ***REMOVED***,
    set: { enumerable: true ***REMOVED***,
    keys: { enumerable: true ***REMOVED***,
    values: { enumerable: true ***REMOVED***,
    entries: { enumerable: true ***REMOVED***,
    forEach: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "FormData", configurable: true ***REMOVED***,
    [Symbol.iterator]: { value: FormData.prototype.entries, configurable: true, writable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = FormData;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: FormData
  ***REMOVED***);
***REMOVED***;

const Impl = require("../xhr/FormData-impl.js");
