// jsdom/lib/jsdom/living/generated/HTMLOptionElement.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const HTMLConstructor_helpers_html_constructor = require("../helpers/html-constructor.js").HTMLConstructor;
const ceReactionsPreSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPreSteps;
const ceReactionsPostSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPostSteps;
const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;
const HTMLElement = require("./HTMLElement.js");

const interfaceName = "HTMLOptionElement";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'HTMLOptionElement'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["HTMLOptionElement"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor HTMLOptionElement is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {
  HTMLElement._internalSetup(wrapper, globalObject);
***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***

  if (globalObject.HTMLElement === undefined) {
    throw new Error("Internal error: attempting to evaluate HTMLOptionElement before HTMLElement");
  ***REMOVED***
  class HTMLOptionElement extends globalObject.HTMLElement {
    constructor() {
      return HTMLConstructor_helpers_html_constructor(globalObject, interfaceName, new.target);
    ***REMOVED***

    get disabled() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get disabled' called on an object that is not a valid instance of HTMLOptionElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].hasAttributeNS(null, "disabled");
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set disabled(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set disabled' called on an object that is not a valid instance of HTMLOptionElement.");
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'disabled' property on 'HTMLOptionElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        if (V) {
          esValue[implSymbol].setAttributeNS(null, "disabled", "");
        ***REMOVED*** else {
          esValue[implSymbol].removeAttributeNS(null, "disabled");
        ***REMOVED***
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get form() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get form' called on an object that is not a valid instance of HTMLOptionElement.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["form"]);
    ***REMOVED***

    get label() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get label' called on an object that is not a valid instance of HTMLOptionElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol]["label"];
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set label(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set label' called on an object that is not a valid instance of HTMLOptionElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'label' property on 'HTMLOptionElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol]["label"] = V;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get defaultSelected() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get defaultSelected' called on an object that is not a valid instance of HTMLOptionElement."
        );
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].hasAttributeNS(null, "selected");
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set defaultSelected(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set defaultSelected' called on an object that is not a valid instance of HTMLOptionElement."
        );
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'defaultSelected' property on 'HTMLOptionElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        if (V) {
          esValue[implSymbol].setAttributeNS(null, "selected", "");
        ***REMOVED*** else {
          esValue[implSymbol].removeAttributeNS(null, "selected");
        ***REMOVED***
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get selected() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get selected' called on an object that is not a valid instance of HTMLOptionElement.");
      ***REMOVED***

      return esValue[implSymbol]["selected"];
    ***REMOVED***

    set selected(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set selected' called on an object that is not a valid instance of HTMLOptionElement.");
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'selected' property on 'HTMLOptionElement': The provided value"
      ***REMOVED***);

      esValue[implSymbol]["selected"] = V;
    ***REMOVED***

    get value() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get value' called on an object that is not a valid instance of HTMLOptionElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol]["value"];
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set value(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set value' called on an object that is not a valid instance of HTMLOptionElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'value' property on 'HTMLOptionElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol]["value"] = V;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get text() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get text' called on an object that is not a valid instance of HTMLOptionElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol]["text"];
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set text(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set text' called on an object that is not a valid instance of HTMLOptionElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'text' property on 'HTMLOptionElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol]["text"] = V;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get index() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get index' called on an object that is not a valid instance of HTMLOptionElement.");
      ***REMOVED***

      return esValue[implSymbol]["index"];
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(HTMLOptionElement.prototype, {
    disabled: { enumerable: true ***REMOVED***,
    form: { enumerable: true ***REMOVED***,
    label: { enumerable: true ***REMOVED***,
    defaultSelected: { enumerable: true ***REMOVED***,
    selected: { enumerable: true ***REMOVED***,
    value: { enumerable: true ***REMOVED***,
    text: { enumerable: true ***REMOVED***,
    index: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "HTMLOptionElement", configurable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = HTMLOptionElement;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: HTMLOptionElement
  ***REMOVED***);
***REMOVED***;

const Impl = require("../nodes/HTMLOptionElement-impl.js");
