// jsdom/lib/jsdom/living/generated/HTMLInputElement.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const HTMLConstructor_helpers_html_constructor = require("../helpers/html-constructor.js").HTMLConstructor;
const SelectionMode = require("./SelectionMode.js");
const ceReactionsPreSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPreSteps;
const ceReactionsPostSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPostSteps;
const FileList = require("./FileList.js");
const parseURLToResultingURLRecord_helpers_document_base_url =
  require("../helpers/document-base-url.js").parseURLToResultingURLRecord;
const serializeURLwhatwg_url = require("whatwg-url").serializeURL;
const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;
const HTMLElement = require("./HTMLElement.js");

const interfaceName = "HTMLInputElement";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'HTMLInputElement'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["HTMLInputElement"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor HTMLInputElement is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {
  HTMLElement._internalSetup(wrapper, globalObject);
***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***

  if (globalObject.HTMLElement === undefined) {
    throw new Error("Internal error: attempting to evaluate HTMLInputElement before HTMLElement");
  ***REMOVED***
  class HTMLInputElement extends globalObject.HTMLElement {
    constructor() {
      return HTMLConstructor_helpers_html_constructor(globalObject, interfaceName, new.target);
    ***REMOVED***

    stepUp() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'stepUp' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        if (curArg !== undefined) {
          curArg = conversions["long"](curArg, {
            context: "Failed to execute 'stepUp' on 'HTMLInputElement': parameter 1"
          ***REMOVED***);
        ***REMOVED*** else {
          curArg = 1;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].stepUp(...args);
    ***REMOVED***

    stepDown() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'stepDown' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        if (curArg !== undefined) {
          curArg = conversions["long"](curArg, {
            context: "Failed to execute 'stepDown' on 'HTMLInputElement': parameter 1"
          ***REMOVED***);
        ***REMOVED*** else {
          curArg = 1;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].stepDown(...args);
    ***REMOVED***

    checkValidity() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'checkValidity' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      return esValue[implSymbol].checkValidity();
    ***REMOVED***

    reportValidity() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'reportValidity' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      return esValue[implSymbol].reportValidity();
    ***REMOVED***

    setCustomValidity(error) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError(
          "'setCustomValidity' called on an object that is not a valid instance of HTMLInputElement."
        );
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'setCustomValidity' on 'HTMLInputElement': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'setCustomValidity' on 'HTMLInputElement': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].setCustomValidity(...args);
    ***REMOVED***

    select() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'select' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      return esValue[implSymbol].select();
    ***REMOVED***

    setRangeText(replacement) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'setRangeText' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'setRangeText' on 'HTMLInputElement': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      switch (arguments.length) {
        case 1:
          {
            let curArg = arguments[0];
            curArg = conversions["DOMString"](curArg, {
              context: "Failed to execute 'setRangeText' on 'HTMLInputElement': parameter 1"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          break;
        case 2:
          throw new TypeError(
            "Failed to execute 'setRangeText' on 'HTMLInputElement': only " + arguments.length + " arguments present."
          );
          break;
        case 3:
          {
            let curArg = arguments[0];
            curArg = conversions["DOMString"](curArg, {
              context: "Failed to execute 'setRangeText' on 'HTMLInputElement': parameter 1"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[1];
            curArg = conversions["unsigned long"](curArg, {
              context: "Failed to execute 'setRangeText' on 'HTMLInputElement': parameter 2"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[2];
            curArg = conversions["unsigned long"](curArg, {
              context: "Failed to execute 'setRangeText' on 'HTMLInputElement': parameter 3"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          break;
        default:
          {
            let curArg = arguments[0];
            curArg = conversions["DOMString"](curArg, {
              context: "Failed to execute 'setRangeText' on 'HTMLInputElement': parameter 1"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[1];
            curArg = conversions["unsigned long"](curArg, {
              context: "Failed to execute 'setRangeText' on 'HTMLInputElement': parameter 2"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[2];
            curArg = conversions["unsigned long"](curArg, {
              context: "Failed to execute 'setRangeText' on 'HTMLInputElement': parameter 3"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[3];
            if (curArg !== undefined) {
              curArg = SelectionMode.convert(curArg, {
                context: "Failed to execute 'setRangeText' on 'HTMLInputElement': parameter 4"
              ***REMOVED***);
            ***REMOVED*** else {
              curArg = "preserve";
            ***REMOVED***
            args.push(curArg);
          ***REMOVED***
      ***REMOVED***
      return esValue[implSymbol].setRangeText(...args);
    ***REMOVED***

    setSelectionRange(start, end) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError(
          "'setSelectionRange' called on an object that is not a valid instance of HTMLInputElement."
        );
      ***REMOVED***

      if (arguments.length < 2) {
        throw new TypeError(
          "Failed to execute 'setSelectionRange' on 'HTMLInputElement': 2 arguments required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["unsigned long"](curArg, {
          context: "Failed to execute 'setSelectionRange' on 'HTMLInputElement': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        curArg = conversions["unsigned long"](curArg, {
          context: "Failed to execute 'setSelectionRange' on 'HTMLInputElement': parameter 2"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[2];
        if (curArg !== undefined) {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'setSelectionRange' on 'HTMLInputElement': parameter 3"
          ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].setSelectionRange(...args);
    ***REMOVED***

    get accept() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get accept' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "accept");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set accept(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set accept' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'accept' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "accept", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get alt() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get alt' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "alt");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set alt(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set alt' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'alt' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "alt", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get autocomplete() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get autocomplete' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "autocomplete");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set autocomplete(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set autocomplete' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'autocomplete' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "autocomplete", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get autofocus() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get autofocus' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].hasAttributeNS(null, "autofocus");
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set autofocus(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set autofocus' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'autofocus' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        if (V) {
          esValue[implSymbol].setAttributeNS(null, "autofocus", "");
        ***REMOVED*** else {
          esValue[implSymbol].removeAttributeNS(null, "autofocus");
        ***REMOVED***
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get defaultChecked() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get defaultChecked' called on an object that is not a valid instance of HTMLInputElement."
        );
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].hasAttributeNS(null, "checked");
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set defaultChecked(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set defaultChecked' called on an object that is not a valid instance of HTMLInputElement."
        );
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'defaultChecked' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        if (V) {
          esValue[implSymbol].setAttributeNS(null, "checked", "");
        ***REMOVED*** else {
          esValue[implSymbol].removeAttributeNS(null, "checked");
        ***REMOVED***
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get checked() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get checked' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      return esValue[implSymbol]["checked"];
    ***REMOVED***

    set checked(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set checked' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'checked' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      esValue[implSymbol]["checked"] = V;
    ***REMOVED***

    get dirName() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get dirName' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "dirname");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set dirName(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set dirName' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'dirName' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "dirname", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get disabled() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get disabled' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].hasAttributeNS(null, "disabled");
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set disabled(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set disabled' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'disabled' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        if (V) {
          esValue[implSymbol].setAttributeNS(null, "disabled", "");
        ***REMOVED*** else {
          esValue[implSymbol].removeAttributeNS(null, "disabled");
        ***REMOVED***
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get form() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get form' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["form"]);
    ***REMOVED***

    get files() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get files' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["files"]);
    ***REMOVED***

    set files(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set files' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      if (V === null || V === undefined) {
        V = null;
      ***REMOVED*** else {
        V = FileList.convert(V, {
          context: "Failed to set the 'files' property on 'HTMLInputElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["files"] = V;
    ***REMOVED***

    get formNoValidate() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get formNoValidate' called on an object that is not a valid instance of HTMLInputElement."
        );
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].hasAttributeNS(null, "formnovalidate");
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set formNoValidate(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set formNoValidate' called on an object that is not a valid instance of HTMLInputElement."
        );
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'formNoValidate' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        if (V) {
          esValue[implSymbol].setAttributeNS(null, "formnovalidate", "");
        ***REMOVED*** else {
          esValue[implSymbol].removeAttributeNS(null, "formnovalidate");
        ***REMOVED***
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get formTarget() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get formTarget' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "formtarget");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set formTarget(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set formTarget' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'formTarget' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "formtarget", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get indeterminate() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get indeterminate' called on an object that is not a valid instance of HTMLInputElement."
        );
      ***REMOVED***

      return esValue[implSymbol]["indeterminate"];
    ***REMOVED***

    set indeterminate(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set indeterminate' called on an object that is not a valid instance of HTMLInputElement."
        );
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'indeterminate' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      esValue[implSymbol]["indeterminate"] = V;
    ***REMOVED***

    get inputMode() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get inputMode' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "inputmode");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set inputMode(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set inputMode' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'inputMode' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "inputmode", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get list() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get list' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["list"]);
    ***REMOVED***

    get max() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get max' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "max");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set max(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set max' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'max' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "max", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get maxLength() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get maxLength' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol]["maxLength"];
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set maxLength(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set maxLength' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["long"](V, {
        context: "Failed to set the 'maxLength' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol]["maxLength"] = V;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get min() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get min' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "min");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set min(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set min' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'min' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "min", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get minLength() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get minLength' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol]["minLength"];
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set minLength(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set minLength' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["long"](V, {
        context: "Failed to set the 'minLength' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol]["minLength"] = V;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get multiple() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get multiple' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].hasAttributeNS(null, "multiple");
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set multiple(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set multiple' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'multiple' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        if (V) {
          esValue[implSymbol].setAttributeNS(null, "multiple", "");
        ***REMOVED*** else {
          esValue[implSymbol].removeAttributeNS(null, "multiple");
        ***REMOVED***
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get name() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get name' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "name");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set name(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set name' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'name' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "name", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get pattern() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get pattern' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "pattern");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set pattern(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set pattern' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'pattern' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "pattern", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get placeholder() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get placeholder' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "placeholder");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set placeholder(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set placeholder' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'placeholder' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "placeholder", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get readOnly() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get readOnly' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].hasAttributeNS(null, "readonly");
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set readOnly(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set readOnly' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'readOnly' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        if (V) {
          esValue[implSymbol].setAttributeNS(null, "readonly", "");
        ***REMOVED*** else {
          esValue[implSymbol].removeAttributeNS(null, "readonly");
        ***REMOVED***
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get required() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get required' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].hasAttributeNS(null, "required");
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set required(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set required' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'required' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        if (V) {
          esValue[implSymbol].setAttributeNS(null, "required", "");
        ***REMOVED*** else {
          esValue[implSymbol].removeAttributeNS(null, "required");
        ***REMOVED***
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get size() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get size' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol]["size"];
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set size(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set size' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["unsigned long"](V, {
        context: "Failed to set the 'size' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol]["size"] = V;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get src() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get src' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "src");
        if (value === null) {
          return "";
        ***REMOVED***
        const urlRecord = parseURLToResultingURLRecord_helpers_document_base_url(
          value,
          esValue[implSymbol]._ownerDocument
        );
        if (urlRecord !== null) {
          return serializeURLwhatwg_url(urlRecord);
        ***REMOVED***
        return conversions.USVString(value);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set src(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set src' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["USVString"](V, {
        context: "Failed to set the 'src' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "src", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get step() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get step' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "step");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set step(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set step' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'step' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "step", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get type() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get type' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol]["type"];
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set type(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set type' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'type' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol]["type"] = V;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get defaultValue() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get defaultValue' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "value");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set defaultValue(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set defaultValue' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'defaultValue' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "value", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get value() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get value' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol]["value"];
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set value(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set value' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'value' property on 'HTMLInputElement': The provided value",
        treatNullAsEmptyString: true
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol]["value"] = V;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get valueAsDate() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get valueAsDate' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      return esValue[implSymbol]["valueAsDate"];
    ***REMOVED***

    set valueAsDate(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set valueAsDate' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      if (V === null || V === undefined) {
        V = null;
      ***REMOVED*** else {
        V = conversions["object"](V, {
          context: "Failed to set the 'valueAsDate' property on 'HTMLInputElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["valueAsDate"] = V;
    ***REMOVED***

    get valueAsNumber() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get valueAsNumber' called on an object that is not a valid instance of HTMLInputElement."
        );
      ***REMOVED***

      return esValue[implSymbol]["valueAsNumber"];
    ***REMOVED***

    set valueAsNumber(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set valueAsNumber' called on an object that is not a valid instance of HTMLInputElement."
        );
      ***REMOVED***

      V = conversions["unrestricted double"](V, {
        context: "Failed to set the 'valueAsNumber' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      esValue[implSymbol]["valueAsNumber"] = V;
    ***REMOVED***

    get willValidate() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get willValidate' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      return esValue[implSymbol]["willValidate"];
    ***REMOVED***

    get validity() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get validity' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["validity"]);
    ***REMOVED***

    get validationMessage() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get validationMessage' called on an object that is not a valid instance of HTMLInputElement."
        );
      ***REMOVED***

      return esValue[implSymbol]["validationMessage"];
    ***REMOVED***

    get labels() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get labels' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["labels"]);
    ***REMOVED***

    get selectionStart() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get selectionStart' called on an object that is not a valid instance of HTMLInputElement."
        );
      ***REMOVED***

      return esValue[implSymbol]["selectionStart"];
    ***REMOVED***

    set selectionStart(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set selectionStart' called on an object that is not a valid instance of HTMLInputElement."
        );
      ***REMOVED***

      if (V === null || V === undefined) {
        V = null;
      ***REMOVED*** else {
        V = conversions["unsigned long"](V, {
          context: "Failed to set the 'selectionStart' property on 'HTMLInputElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["selectionStart"] = V;
    ***REMOVED***

    get selectionEnd() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get selectionEnd' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      return esValue[implSymbol]["selectionEnd"];
    ***REMOVED***

    set selectionEnd(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set selectionEnd' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      if (V === null || V === undefined) {
        V = null;
      ***REMOVED*** else {
        V = conversions["unsigned long"](V, {
          context: "Failed to set the 'selectionEnd' property on 'HTMLInputElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["selectionEnd"] = V;
    ***REMOVED***

    get selectionDirection() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get selectionDirection' called on an object that is not a valid instance of HTMLInputElement."
        );
      ***REMOVED***

      return esValue[implSymbol]["selectionDirection"];
    ***REMOVED***

    set selectionDirection(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set selectionDirection' called on an object that is not a valid instance of HTMLInputElement."
        );
      ***REMOVED***

      if (V === null || V === undefined) {
        V = null;
      ***REMOVED*** else {
        V = conversions["DOMString"](V, {
          context: "Failed to set the 'selectionDirection' property on 'HTMLInputElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["selectionDirection"] = V;
    ***REMOVED***

    get align() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get align' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "align");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set align(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set align' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'align' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "align", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get useMap() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get useMap' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "usemap");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set useMap(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set useMap' called on an object that is not a valid instance of HTMLInputElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'useMap' property on 'HTMLInputElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "usemap", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(HTMLInputElement.prototype, {
    stepUp: { enumerable: true ***REMOVED***,
    stepDown: { enumerable: true ***REMOVED***,
    checkValidity: { enumerable: true ***REMOVED***,
    reportValidity: { enumerable: true ***REMOVED***,
    setCustomValidity: { enumerable: true ***REMOVED***,
    select: { enumerable: true ***REMOVED***,
    setRangeText: { enumerable: true ***REMOVED***,
    setSelectionRange: { enumerable: true ***REMOVED***,
    accept: { enumerable: true ***REMOVED***,
    alt: { enumerable: true ***REMOVED***,
    autocomplete: { enumerable: true ***REMOVED***,
    autofocus: { enumerable: true ***REMOVED***,
    defaultChecked: { enumerable: true ***REMOVED***,
    checked: { enumerable: true ***REMOVED***,
    dirName: { enumerable: true ***REMOVED***,
    disabled: { enumerable: true ***REMOVED***,
    form: { enumerable: true ***REMOVED***,
    files: { enumerable: true ***REMOVED***,
    formNoValidate: { enumerable: true ***REMOVED***,
    formTarget: { enumerable: true ***REMOVED***,
    indeterminate: { enumerable: true ***REMOVED***,
    inputMode: { enumerable: true ***REMOVED***,
    list: { enumerable: true ***REMOVED***,
    max: { enumerable: true ***REMOVED***,
    maxLength: { enumerable: true ***REMOVED***,
    min: { enumerable: true ***REMOVED***,
    minLength: { enumerable: true ***REMOVED***,
    multiple: { enumerable: true ***REMOVED***,
    name: { enumerable: true ***REMOVED***,
    pattern: { enumerable: true ***REMOVED***,
    placeholder: { enumerable: true ***REMOVED***,
    readOnly: { enumerable: true ***REMOVED***,
    required: { enumerable: true ***REMOVED***,
    size: { enumerable: true ***REMOVED***,
    src: { enumerable: true ***REMOVED***,
    step: { enumerable: true ***REMOVED***,
    type: { enumerable: true ***REMOVED***,
    defaultValue: { enumerable: true ***REMOVED***,
    value: { enumerable: true ***REMOVED***,
    valueAsDate: { enumerable: true ***REMOVED***,
    valueAsNumber: { enumerable: true ***REMOVED***,
    willValidate: { enumerable: true ***REMOVED***,
    validity: { enumerable: true ***REMOVED***,
    validationMessage: { enumerable: true ***REMOVED***,
    labels: { enumerable: true ***REMOVED***,
    selectionStart: { enumerable: true ***REMOVED***,
    selectionEnd: { enumerable: true ***REMOVED***,
    selectionDirection: { enumerable: true ***REMOVED***,
    align: { enumerable: true ***REMOVED***,
    useMap: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "HTMLInputElement", configurable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = HTMLInputElement;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: HTMLInputElement
  ***REMOVED***);
***REMOVED***;

const Impl = require("../nodes/HTMLInputElement-impl.js");
