// jsdom/lib/jsdom/living/generated/HTMLMediaElement.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const TextTrackKind = require("./TextTrackKind.js");
const parseURLToResultingURLRecord_helpers_document_base_url =
  require("../helpers/document-base-url.js").parseURLToResultingURLRecord;
const serializeURLwhatwg_url = require("whatwg-url").serializeURL;
const ceReactionsPreSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPreSteps;
const ceReactionsPostSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPostSteps;
const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;
const HTMLElement = require("./HTMLElement.js");

const interfaceName = "HTMLMediaElement";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'HTMLMediaElement'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["HTMLMediaElement"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor HTMLMediaElement is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {
  HTMLElement._internalSetup(wrapper, globalObject);
***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***

  if (globalObject.HTMLElement === undefined) {
    throw new Error("Internal error: attempting to evaluate HTMLMediaElement before HTMLElement");
  ***REMOVED***
  class HTMLMediaElement extends globalObject.HTMLElement {
    constructor() {
      throw new TypeError("Illegal constructor");
    ***REMOVED***

    load() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'load' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return esValue[implSymbol].load();
    ***REMOVED***

    canPlayType(type) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'canPlayType' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'canPlayType' on 'HTMLMediaElement': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'canPlayType' on 'HTMLMediaElement': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].canPlayType(...args));
    ***REMOVED***

    play() {
      try {
        const esValue = this !== null && this !== undefined ? this : globalObject;
        if (!exports.is(esValue)) {
          throw new TypeError("'play' called on an object that is not a valid instance of HTMLMediaElement.");
        ***REMOVED***

        return utils.tryWrapperForImpl(esValue[implSymbol].play());
      ***REMOVED*** catch (e) {
        return Promise.reject(e);
      ***REMOVED***
    ***REMOVED***

    pause() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'pause' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return esValue[implSymbol].pause();
    ***REMOVED***

    addTextTrack(kind) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'addTextTrack' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'addTextTrack' on 'HTMLMediaElement': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = TextTrackKind.convert(curArg, {
          context: "Failed to execute 'addTextTrack' on 'HTMLMediaElement': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        if (curArg !== undefined) {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'addTextTrack' on 'HTMLMediaElement': parameter 2"
          ***REMOVED***);
        ***REMOVED*** else {
          curArg = "";
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[2];
        if (curArg !== undefined) {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'addTextTrack' on 'HTMLMediaElement': parameter 3"
          ***REMOVED***);
        ***REMOVED*** else {
          curArg = "";
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].addTextTrack(...args));
    ***REMOVED***

    get src() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get src' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "src");
        if (value === null) {
          return "";
        ***REMOVED***
        const urlRecord = parseURLToResultingURLRecord_helpers_document_base_url(
          value,
          esValue[implSymbol]._ownerDocument
        );
        if (urlRecord !== null) {
          return serializeURLwhatwg_url(urlRecord);
        ***REMOVED***
        return conversions.USVString(value);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set src(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set src' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      V = conversions["USVString"](V, {
        context: "Failed to set the 'src' property on 'HTMLMediaElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "src", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get currentSrc() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get currentSrc' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return esValue[implSymbol]["currentSrc"];
    ***REMOVED***

    get crossOrigin() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get crossOrigin' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "crossorigin");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set crossOrigin(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set crossOrigin' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      if (V === null || V === undefined) {
        V = null;
      ***REMOVED*** else {
        V = conversions["DOMString"](V, {
          context: "Failed to set the 'crossOrigin' property on 'HTMLMediaElement': The provided value"
        ***REMOVED***);
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "crossorigin", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get networkState() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get networkState' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return esValue[implSymbol]["networkState"];
    ***REMOVED***

    get preload() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get preload' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "preload");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set preload(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set preload' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'preload' property on 'HTMLMediaElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "preload", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get buffered() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get buffered' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["buffered"]);
    ***REMOVED***

    get readyState() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get readyState' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return esValue[implSymbol]["readyState"];
    ***REMOVED***

    get seeking() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get seeking' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return esValue[implSymbol]["seeking"];
    ***REMOVED***

    get currentTime() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get currentTime' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return esValue[implSymbol]["currentTime"];
    ***REMOVED***

    set currentTime(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set currentTime' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      V = conversions["double"](V, {
        context: "Failed to set the 'currentTime' property on 'HTMLMediaElement': The provided value"
      ***REMOVED***);

      esValue[implSymbol]["currentTime"] = V;
    ***REMOVED***

    get duration() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get duration' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return esValue[implSymbol]["duration"];
    ***REMOVED***

    get paused() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get paused' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return esValue[implSymbol]["paused"];
    ***REMOVED***

    get defaultPlaybackRate() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get defaultPlaybackRate' called on an object that is not a valid instance of HTMLMediaElement."
        );
      ***REMOVED***

      return esValue[implSymbol]["defaultPlaybackRate"];
    ***REMOVED***

    set defaultPlaybackRate(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set defaultPlaybackRate' called on an object that is not a valid instance of HTMLMediaElement."
        );
      ***REMOVED***

      V = conversions["double"](V, {
        context: "Failed to set the 'defaultPlaybackRate' property on 'HTMLMediaElement': The provided value"
      ***REMOVED***);

      esValue[implSymbol]["defaultPlaybackRate"] = V;
    ***REMOVED***

    get playbackRate() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get playbackRate' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return esValue[implSymbol]["playbackRate"];
    ***REMOVED***

    set playbackRate(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set playbackRate' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      V = conversions["double"](V, {
        context: "Failed to set the 'playbackRate' property on 'HTMLMediaElement': The provided value"
      ***REMOVED***);

      esValue[implSymbol]["playbackRate"] = V;
    ***REMOVED***

    get played() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get played' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["played"]);
    ***REMOVED***

    get seekable() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get seekable' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["seekable"]);
    ***REMOVED***

    get ended() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get ended' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return esValue[implSymbol]["ended"];
    ***REMOVED***

    get autoplay() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get autoplay' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].hasAttributeNS(null, "autoplay");
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set autoplay(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set autoplay' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'autoplay' property on 'HTMLMediaElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        if (V) {
          esValue[implSymbol].setAttributeNS(null, "autoplay", "");
        ***REMOVED*** else {
          esValue[implSymbol].removeAttributeNS(null, "autoplay");
        ***REMOVED***
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get loop() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get loop' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].hasAttributeNS(null, "loop");
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set loop(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set loop' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'loop' property on 'HTMLMediaElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        if (V) {
          esValue[implSymbol].setAttributeNS(null, "loop", "");
        ***REMOVED*** else {
          esValue[implSymbol].removeAttributeNS(null, "loop");
        ***REMOVED***
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get controls() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get controls' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].hasAttributeNS(null, "controls");
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set controls(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set controls' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'controls' property on 'HTMLMediaElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        if (V) {
          esValue[implSymbol].setAttributeNS(null, "controls", "");
        ***REMOVED*** else {
          esValue[implSymbol].removeAttributeNS(null, "controls");
        ***REMOVED***
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get volume() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get volume' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return esValue[implSymbol]["volume"];
    ***REMOVED***

    set volume(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set volume' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      V = conversions["double"](V, {
        context: "Failed to set the 'volume' property on 'HTMLMediaElement': The provided value"
      ***REMOVED***);

      esValue[implSymbol]["volume"] = V;
    ***REMOVED***

    get muted() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get muted' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return esValue[implSymbol]["muted"];
    ***REMOVED***

    set muted(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set muted' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'muted' property on 'HTMLMediaElement': The provided value"
      ***REMOVED***);

      esValue[implSymbol]["muted"] = V;
    ***REMOVED***

    get defaultMuted() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get defaultMuted' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].hasAttributeNS(null, "muted");
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set defaultMuted(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set defaultMuted' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'defaultMuted' property on 'HTMLMediaElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        if (V) {
          esValue[implSymbol].setAttributeNS(null, "muted", "");
        ***REMOVED*** else {
          esValue[implSymbol].removeAttributeNS(null, "muted");
        ***REMOVED***
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get audioTracks() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get audioTracks' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return utils.getSameObject(this, "audioTracks", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["audioTracks"]);
      ***REMOVED***);
    ***REMOVED***

    get videoTracks() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get videoTracks' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return utils.getSameObject(this, "videoTracks", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["videoTracks"]);
      ***REMOVED***);
    ***REMOVED***

    get textTracks() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get textTracks' called on an object that is not a valid instance of HTMLMediaElement.");
      ***REMOVED***

      return utils.getSameObject(this, "textTracks", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["textTracks"]);
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(HTMLMediaElement.prototype, {
    load: { enumerable: true ***REMOVED***,
    canPlayType: { enumerable: true ***REMOVED***,
    play: { enumerable: true ***REMOVED***,
    pause: { enumerable: true ***REMOVED***,
    addTextTrack: { enumerable: true ***REMOVED***,
    src: { enumerable: true ***REMOVED***,
    currentSrc: { enumerable: true ***REMOVED***,
    crossOrigin: { enumerable: true ***REMOVED***,
    networkState: { enumerable: true ***REMOVED***,
    preload: { enumerable: true ***REMOVED***,
    buffered: { enumerable: true ***REMOVED***,
    readyState: { enumerable: true ***REMOVED***,
    seeking: { enumerable: true ***REMOVED***,
    currentTime: { enumerable: true ***REMOVED***,
    duration: { enumerable: true ***REMOVED***,
    paused: { enumerable: true ***REMOVED***,
    defaultPlaybackRate: { enumerable: true ***REMOVED***,
    playbackRate: { enumerable: true ***REMOVED***,
    played: { enumerable: true ***REMOVED***,
    seekable: { enumerable: true ***REMOVED***,
    ended: { enumerable: true ***REMOVED***,
    autoplay: { enumerable: true ***REMOVED***,
    loop: { enumerable: true ***REMOVED***,
    controls: { enumerable: true ***REMOVED***,
    volume: { enumerable: true ***REMOVED***,
    muted: { enumerable: true ***REMOVED***,
    defaultMuted: { enumerable: true ***REMOVED***,
    audioTracks: { enumerable: true ***REMOVED***,
    videoTracks: { enumerable: true ***REMOVED***,
    textTracks: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "HTMLMediaElement", configurable: true ***REMOVED***,
    NETWORK_EMPTY: { value: 0, enumerable: true ***REMOVED***,
    NETWORK_IDLE: { value: 1, enumerable: true ***REMOVED***,
    NETWORK_LOADING: { value: 2, enumerable: true ***REMOVED***,
    NETWORK_NO_SOURCE: { value: 3, enumerable: true ***REMOVED***,
    HAVE_NOTHING: { value: 0, enumerable: true ***REMOVED***,
    HAVE_METADATA: { value: 1, enumerable: true ***REMOVED***,
    HAVE_CURRENT_DATA: { value: 2, enumerable: true ***REMOVED***,
    HAVE_FUTURE_DATA: { value: 3, enumerable: true ***REMOVED***,
    HAVE_ENOUGH_DATA: { value: 4, enumerable: true ***REMOVED***
  ***REMOVED***);
  Object.defineProperties(HTMLMediaElement, {
    NETWORK_EMPTY: { value: 0, enumerable: true ***REMOVED***,
    NETWORK_IDLE: { value: 1, enumerable: true ***REMOVED***,
    NETWORK_LOADING: { value: 2, enumerable: true ***REMOVED***,
    NETWORK_NO_SOURCE: { value: 3, enumerable: true ***REMOVED***,
    HAVE_NOTHING: { value: 0, enumerable: true ***REMOVED***,
    HAVE_METADATA: { value: 1, enumerable: true ***REMOVED***,
    HAVE_CURRENT_DATA: { value: 2, enumerable: true ***REMOVED***,
    HAVE_FUTURE_DATA: { value: 3, enumerable: true ***REMOVED***,
    HAVE_ENOUGH_DATA: { value: 4, enumerable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = HTMLMediaElement;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: HTMLMediaElement
  ***REMOVED***);
***REMOVED***;

const Impl = require("../nodes/HTMLMediaElement-impl.js");
