// jsdom/lib/jsdom/living/generated/HTMLFrameSetElement.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const HTMLConstructor_helpers_html_constructor = require("../helpers/html-constructor.js").HTMLConstructor;
const ceReactionsPreSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPreSteps;
const ceReactionsPostSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPostSteps;
const EventHandlerNonNull = require("./EventHandlerNonNull.js");
const OnBeforeUnloadEventHandlerNonNull = require("./OnBeforeUnloadEventHandlerNonNull.js");
const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;
const HTMLElement = require("./HTMLElement.js");

const interfaceName = "HTMLFrameSetElement";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'HTMLFrameSetElement'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["HTMLFrameSetElement"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor HTMLFrameSetElement is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {
  HTMLElement._internalSetup(wrapper, globalObject);
***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***

  if (globalObject.HTMLElement === undefined) {
    throw new Error("Internal error: attempting to evaluate HTMLFrameSetElement before HTMLElement");
  ***REMOVED***
  class HTMLFrameSetElement extends globalObject.HTMLElement {
    constructor() {
      return HTMLConstructor_helpers_html_constructor(globalObject, interfaceName, new.target);
    ***REMOVED***

    get cols() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get cols' called on an object that is not a valid instance of HTMLFrameSetElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "cols");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set cols(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set cols' called on an object that is not a valid instance of HTMLFrameSetElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'cols' property on 'HTMLFrameSetElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "cols", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get rows() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get rows' called on an object that is not a valid instance of HTMLFrameSetElement.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "rows");
        return value === null ? "" : value;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set rows(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set rows' called on an object that is not a valid instance of HTMLFrameSetElement.");
      ***REMOVED***

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'rows' property on 'HTMLFrameSetElement': The provided value"
      ***REMOVED***);

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol].setAttributeNS(null, "rows", V);
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get onafterprint() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onafterprint' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onafterprint"]);
    ***REMOVED***

    set onafterprint(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onafterprint' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onafterprint' property on 'HTMLFrameSetElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onafterprint"] = V;
    ***REMOVED***

    get onbeforeprint() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onbeforeprint' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onbeforeprint"]);
    ***REMOVED***

    set onbeforeprint(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onbeforeprint' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onbeforeprint' property on 'HTMLFrameSetElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onbeforeprint"] = V;
    ***REMOVED***

    get onbeforeunload() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onbeforeunload' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onbeforeunload"]);
    ***REMOVED***

    set onbeforeunload(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onbeforeunload' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = OnBeforeUnloadEventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onbeforeunload' property on 'HTMLFrameSetElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onbeforeunload"] = V;
    ***REMOVED***

    get onhashchange() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onhashchange' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onhashchange"]);
    ***REMOVED***

    set onhashchange(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onhashchange' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onhashchange' property on 'HTMLFrameSetElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onhashchange"] = V;
    ***REMOVED***

    get onlanguagechange() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onlanguagechange' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onlanguagechange"]);
    ***REMOVED***

    set onlanguagechange(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onlanguagechange' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onlanguagechange' property on 'HTMLFrameSetElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onlanguagechange"] = V;
    ***REMOVED***

    get onmessage() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onmessage' called on an object that is not a valid instance of HTMLFrameSetElement.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onmessage"]);
    ***REMOVED***

    set onmessage(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onmessage' called on an object that is not a valid instance of HTMLFrameSetElement.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onmessage' property on 'HTMLFrameSetElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onmessage"] = V;
    ***REMOVED***

    get onmessageerror() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onmessageerror' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onmessageerror"]);
    ***REMOVED***

    set onmessageerror(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onmessageerror' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onmessageerror' property on 'HTMLFrameSetElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onmessageerror"] = V;
    ***REMOVED***

    get onoffline() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onoffline' called on an object that is not a valid instance of HTMLFrameSetElement.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onoffline"]);
    ***REMOVED***

    set onoffline(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onoffline' called on an object that is not a valid instance of HTMLFrameSetElement.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onoffline' property on 'HTMLFrameSetElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onoffline"] = V;
    ***REMOVED***

    get ononline() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get ononline' called on an object that is not a valid instance of HTMLFrameSetElement.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["ononline"]);
    ***REMOVED***

    set ononline(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set ononline' called on an object that is not a valid instance of HTMLFrameSetElement.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'ononline' property on 'HTMLFrameSetElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["ononline"] = V;
    ***REMOVED***

    get onpagehide() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onpagehide' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onpagehide"]);
    ***REMOVED***

    set onpagehide(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onpagehide' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onpagehide' property on 'HTMLFrameSetElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onpagehide"] = V;
    ***REMOVED***

    get onpageshow() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onpageshow' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onpageshow"]);
    ***REMOVED***

    set onpageshow(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onpageshow' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onpageshow' property on 'HTMLFrameSetElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onpageshow"] = V;
    ***REMOVED***

    get onpopstate() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onpopstate' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onpopstate"]);
    ***REMOVED***

    set onpopstate(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onpopstate' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onpopstate' property on 'HTMLFrameSetElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onpopstate"] = V;
    ***REMOVED***

    get onrejectionhandled() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onrejectionhandled' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onrejectionhandled"]);
    ***REMOVED***

    set onrejectionhandled(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onrejectionhandled' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onrejectionhandled' property on 'HTMLFrameSetElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onrejectionhandled"] = V;
    ***REMOVED***

    get onstorage() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onstorage' called on an object that is not a valid instance of HTMLFrameSetElement.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onstorage"]);
    ***REMOVED***

    set onstorage(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onstorage' called on an object that is not a valid instance of HTMLFrameSetElement.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onstorage' property on 'HTMLFrameSetElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onstorage"] = V;
    ***REMOVED***

    get onunhandledrejection() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onunhandledrejection' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onunhandledrejection"]);
    ***REMOVED***

    set onunhandledrejection(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onunhandledrejection' called on an object that is not a valid instance of HTMLFrameSetElement."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onunhandledrejection' property on 'HTMLFrameSetElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onunhandledrejection"] = V;
    ***REMOVED***

    get onunload() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onunload' called on an object that is not a valid instance of HTMLFrameSetElement.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onunload"]);
    ***REMOVED***

    set onunload(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onunload' called on an object that is not a valid instance of HTMLFrameSetElement.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onunload' property on 'HTMLFrameSetElement': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onunload"] = V;
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(HTMLFrameSetElement.prototype, {
    cols: { enumerable: true ***REMOVED***,
    rows: { enumerable: true ***REMOVED***,
    onafterprint: { enumerable: true ***REMOVED***,
    onbeforeprint: { enumerable: true ***REMOVED***,
    onbeforeunload: { enumerable: true ***REMOVED***,
    onhashchange: { enumerable: true ***REMOVED***,
    onlanguagechange: { enumerable: true ***REMOVED***,
    onmessage: { enumerable: true ***REMOVED***,
    onmessageerror: { enumerable: true ***REMOVED***,
    onoffline: { enumerable: true ***REMOVED***,
    ononline: { enumerable: true ***REMOVED***,
    onpagehide: { enumerable: true ***REMOVED***,
    onpageshow: { enumerable: true ***REMOVED***,
    onpopstate: { enumerable: true ***REMOVED***,
    onrejectionhandled: { enumerable: true ***REMOVED***,
    onstorage: { enumerable: true ***REMOVED***,
    onunhandledrejection: { enumerable: true ***REMOVED***,
    onunload: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "HTMLFrameSetElement", configurable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = HTMLFrameSetElement;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: HTMLFrameSetElement
  ***REMOVED***);
***REMOVED***;

const Impl = require("../nodes/HTMLFrameSetElement-impl.js");
