// jsdom/lib/jsdom/living/generated/Location.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;

const interfaceName = "Location";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'Location'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["Location"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor Location is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {
  Object.defineProperties(
    wrapper,
    Object.getOwnPropertyDescriptors({
      assign(url) {
        const esValue = this !== null && this !== undefined ? this : globalObject;
        if (!exports.is(esValue)) {
          throw new TypeError("'assign' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        if (arguments.length < 1) {
          throw new TypeError(
            "Failed to execute 'assign' on 'Location': 1 argument required, but only " + arguments.length + " present."
          );
        ***REMOVED***
        const args = [];
        {
          let curArg = arguments[0];
          curArg = conversions["USVString"](curArg, {
            context: "Failed to execute 'assign' on 'Location': parameter 1"
          ***REMOVED***);
          args.push(curArg);
        ***REMOVED***
        return esValue[implSymbol].assign(...args);
      ***REMOVED***,
      replace(url) {
        const esValue = this !== null && this !== undefined ? this : globalObject;
        if (!exports.is(esValue)) {
          throw new TypeError("'replace' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        if (arguments.length < 1) {
          throw new TypeError(
            "Failed to execute 'replace' on 'Location': 1 argument required, but only " + arguments.length + " present."
          );
        ***REMOVED***
        const args = [];
        {
          let curArg = arguments[0];
          curArg = conversions["USVString"](curArg, {
            context: "Failed to execute 'replace' on 'Location': parameter 1"
          ***REMOVED***);
          args.push(curArg);
        ***REMOVED***
        return esValue[implSymbol].replace(...args);
      ***REMOVED***,
      reload() {
        const esValue = this !== null && this !== undefined ? this : globalObject;
        if (!exports.is(esValue)) {
          throw new TypeError("'reload' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        return esValue[implSymbol].reload();
      ***REMOVED***,
      get href() {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get href' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        return esValue[implSymbol]["href"];
      ***REMOVED***,
      set href(V) {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set href' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        V = conversions["USVString"](V, {
          context: "Failed to set the 'href' property on 'Location': The provided value"
        ***REMOVED***);

        esValue[implSymbol]["href"] = V;
      ***REMOVED***,
      toString() {
        const esValue = this;
        if (!exports.is(esValue)) {
          throw new TypeError("'toString' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        return esValue[implSymbol]["href"];
      ***REMOVED***,
      get origin() {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get origin' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        return esValue[implSymbol]["origin"];
      ***REMOVED***,
      get protocol() {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get protocol' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        return esValue[implSymbol]["protocol"];
      ***REMOVED***,
      set protocol(V) {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set protocol' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        V = conversions["USVString"](V, {
          context: "Failed to set the 'protocol' property on 'Location': The provided value"
        ***REMOVED***);

        esValue[implSymbol]["protocol"] = V;
      ***REMOVED***,
      get host() {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get host' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        return esValue[implSymbol]["host"];
      ***REMOVED***,
      set host(V) {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set host' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        V = conversions["USVString"](V, {
          context: "Failed to set the 'host' property on 'Location': The provided value"
        ***REMOVED***);

        esValue[implSymbol]["host"] = V;
      ***REMOVED***,
      get hostname() {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get hostname' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        return esValue[implSymbol]["hostname"];
      ***REMOVED***,
      set hostname(V) {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set hostname' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        V = conversions["USVString"](V, {
          context: "Failed to set the 'hostname' property on 'Location': The provided value"
        ***REMOVED***);

        esValue[implSymbol]["hostname"] = V;
      ***REMOVED***,
      get port() {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get port' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        return esValue[implSymbol]["port"];
      ***REMOVED***,
      set port(V) {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set port' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        V = conversions["USVString"](V, {
          context: "Failed to set the 'port' property on 'Location': The provided value"
        ***REMOVED***);

        esValue[implSymbol]["port"] = V;
      ***REMOVED***,
      get pathname() {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get pathname' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        return esValue[implSymbol]["pathname"];
      ***REMOVED***,
      set pathname(V) {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set pathname' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        V = conversions["USVString"](V, {
          context: "Failed to set the 'pathname' property on 'Location': The provided value"
        ***REMOVED***);

        esValue[implSymbol]["pathname"] = V;
      ***REMOVED***,
      get search() {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get search' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        return esValue[implSymbol]["search"];
      ***REMOVED***,
      set search(V) {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set search' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        V = conversions["USVString"](V, {
          context: "Failed to set the 'search' property on 'Location': The provided value"
        ***REMOVED***);

        esValue[implSymbol]["search"] = V;
      ***REMOVED***,
      get hash() {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get hash' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        return esValue[implSymbol]["hash"];
      ***REMOVED***,
      set hash(V) {
        const esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set hash' called on an object that is not a valid instance of Location.");
        ***REMOVED***

        V = conversions["USVString"](V, {
          context: "Failed to set the 'hash' property on 'Location': The provided value"
        ***REMOVED***);

        esValue[implSymbol]["hash"] = V;
      ***REMOVED***
    ***REMOVED***)
  );

  Object.defineProperties(wrapper, {
    assign: { configurable: false, writable: false ***REMOVED***,
    replace: { configurable: false, writable: false ***REMOVED***,
    reload: { configurable: false, writable: false ***REMOVED***,
    href: { configurable: false ***REMOVED***,
    toString: { configurable: false, writable: false ***REMOVED***,
    origin: { configurable: false ***REMOVED***,
    protocol: { configurable: false ***REMOVED***,
    host: { configurable: false ***REMOVED***,
    hostname: { configurable: false ***REMOVED***,
    port: { configurable: false ***REMOVED***,
    pathname: { configurable: false ***REMOVED***,
    search: { configurable: false ***REMOVED***,
    hash: { configurable: false ***REMOVED***
  ***REMOVED***);
***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***
  class Location {
    constructor() {
      throw new TypeError("Illegal constructor");
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(Location.prototype, { [Symbol.toStringTag]: { value: "Location", configurable: true ***REMOVED*** ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = Location;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: Location
  ***REMOVED***);
***REMOVED***;

const Impl = require("../window/Location-impl.js");
