// jsdom/lib/jsdom/living/generated/MimeTypeArray.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;

const interfaceName = "MimeTypeArray";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'MimeTypeArray'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["MimeTypeArray"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor MimeTypeArray is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper = new Proxy(wrapper, proxyHandler);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  let wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper = new Proxy(wrapper, proxyHandler);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***
  class MimeTypeArray {
    constructor() {
      throw new TypeError("Illegal constructor");
    ***REMOVED***

    item(index) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'item' called on an object that is not a valid instance of MimeTypeArray.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'item' on 'MimeTypeArray': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["unsigned long"](curArg, {
          context: "Failed to execute 'item' on 'MimeTypeArray': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].item(...args);
    ***REMOVED***

    namedItem(name) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'namedItem' called on an object that is not a valid instance of MimeTypeArray.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'namedItem' on 'MimeTypeArray': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'namedItem' on 'MimeTypeArray': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].namedItem(...args);
    ***REMOVED***

    get length() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get length' called on an object that is not a valid instance of MimeTypeArray.");
      ***REMOVED***

      return esValue[implSymbol]["length"];
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(MimeTypeArray.prototype, {
    item: { enumerable: true ***REMOVED***,
    namedItem: { enumerable: true ***REMOVED***,
    length: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "MimeTypeArray", configurable: true ***REMOVED***,
    [Symbol.iterator]: { value: Array.prototype[Symbol.iterator], configurable: true, writable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = MimeTypeArray;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: MimeTypeArray
  ***REMOVED***);
***REMOVED***;

const proxyHandler = {
  get(target, P, receiver) {
    if (typeof P === "symbol") {
      return Reflect.get(target, P, receiver);
    ***REMOVED***
    const desc = this.getOwnPropertyDescriptor(target, P);
    if (desc === undefined) {
      const parent = Object.getPrototypeOf(target);
      if (parent === null) {
        return undefined;
      ***REMOVED***
      return Reflect.get(target, P, receiver);
    ***REMOVED***
    if (!desc.get && !desc.set) {
      return desc.value;
    ***REMOVED***
    const getter = desc.get;
    if (getter === undefined) {
      return undefined;
    ***REMOVED***
    return Reflect.apply(getter, receiver, []);
  ***REMOVED***,

  has(target, P) {
    if (typeof P === "symbol") {
      return Reflect.has(target, P);
    ***REMOVED***
    const desc = this.getOwnPropertyDescriptor(target, P);
    if (desc !== undefined) {
      return true;
    ***REMOVED***
    const parent = Object.getPrototypeOf(target);
    if (parent !== null) {
      return Reflect.has(parent, P);
    ***REMOVED***
    return false;
  ***REMOVED***,

  ownKeys(target) {
    const keys = new Set();

    for (const key of target[implSymbol][utils.supportedPropertyIndices]) {
      keys.add(`${key***REMOVED***`);
    ***REMOVED***

    for (const key of Reflect.ownKeys(target)) {
      keys.add(key);
    ***REMOVED***
    return [...keys];
  ***REMOVED***,

  getOwnPropertyDescriptor(target, P) {
    if (typeof P === "symbol") {
      return Reflect.getOwnPropertyDescriptor(target, P);
    ***REMOVED***
    let ignoreNamedProps = false;

    if (utils.isArrayIndexPropName(P)) {
      const index = P >>> 0;
      const indexedValue = target[implSymbol].item(index);
      if (indexedValue !== null) {
        return {
          writable: false,
          enumerable: true,
          configurable: true,
          value: utils.tryWrapperForImpl(indexedValue)
        ***REMOVED***;
      ***REMOVED***
      ignoreNamedProps = true;
    ***REMOVED***

    return Reflect.getOwnPropertyDescriptor(target, P);
  ***REMOVED***,

  set(target, P, V, receiver) {
    if (typeof P === "symbol") {
      return Reflect.set(target, P, V, receiver);
    ***REMOVED***
    // The `receiver` argument refers to the Proxy exotic object or an object
    // that inherits from it, whereas `target` refers to the Proxy target:
    if (target[implSymbol][utils.wrapperSymbol] === receiver) {
    ***REMOVED***
    let ownDesc;

    if (utils.isArrayIndexPropName(P)) {
      const index = P >>> 0;
      const indexedValue = target[implSymbol].item(index);
      if (indexedValue !== null) {
        ownDesc = {
          writable: false,
          enumerable: true,
          configurable: true,
          value: utils.tryWrapperForImpl(indexedValue)
        ***REMOVED***;
      ***REMOVED***
    ***REMOVED***

    if (ownDesc === undefined) {
      ownDesc = Reflect.getOwnPropertyDescriptor(target, P);
    ***REMOVED***
    if (ownDesc === undefined) {
      const parent = Reflect.getPrototypeOf(target);
      if (parent !== null) {
        return Reflect.set(parent, P, V, receiver);
      ***REMOVED***
      ownDesc = { writable: true, enumerable: true, configurable: true, value: undefined ***REMOVED***;
    ***REMOVED***
    if (!ownDesc.writable) {
      return false;
    ***REMOVED***
    if (!utils.isObject(receiver)) {
      return false;
    ***REMOVED***
    const existingDesc = Reflect.getOwnPropertyDescriptor(receiver, P);
    let valueDesc;
    if (existingDesc !== undefined) {
      if (existingDesc.get || existingDesc.set) {
        return false;
      ***REMOVED***
      if (!existingDesc.writable) {
        return false;
      ***REMOVED***
      valueDesc = { value: V ***REMOVED***;
    ***REMOVED*** else {
      valueDesc = { writable: true, enumerable: true, configurable: true, value: V ***REMOVED***;
    ***REMOVED***
    return Reflect.defineProperty(receiver, P, valueDesc);
  ***REMOVED***,

  defineProperty(target, P, desc) {
    if (typeof P === "symbol") {
      return Reflect.defineProperty(target, P, desc);
    ***REMOVED***

    if (utils.isArrayIndexPropName(P)) {
      return false;
    ***REMOVED***

    return Reflect.defineProperty(target, P, desc);
  ***REMOVED***,

  deleteProperty(target, P) {
    if (typeof P === "symbol") {
      return Reflect.deleteProperty(target, P);
    ***REMOVED***

    if (utils.isArrayIndexPropName(P)) {
      const index = P >>> 0;
      return !(target[implSymbol].item(index) !== null);
    ***REMOVED***

    return Reflect.deleteProperty(target, P);
  ***REMOVED***,

  preventExtensions() {
    return false;
  ***REMOVED***
***REMOVED***;

const Impl = require("../navigator/MimeTypeArray-impl.js");
