// jsdom/lib/jsdom/living/generated/Navigator.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;

const interfaceName = "Navigator";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'Navigator'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["Navigator"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor Navigator is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***
  class Navigator {
    constructor() {
      throw new TypeError("Illegal constructor");
    ***REMOVED***

    javaEnabled() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'javaEnabled' called on an object that is not a valid instance of Navigator.");
      ***REMOVED***

      return esValue[implSymbol].javaEnabled();
    ***REMOVED***

    get appCodeName() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get appCodeName' called on an object that is not a valid instance of Navigator.");
      ***REMOVED***

      return esValue[implSymbol]["appCodeName"];
    ***REMOVED***

    get appName() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get appName' called on an object that is not a valid instance of Navigator.");
      ***REMOVED***

      return esValue[implSymbol]["appName"];
    ***REMOVED***

    get appVersion() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get appVersion' called on an object that is not a valid instance of Navigator.");
      ***REMOVED***

      return esValue[implSymbol]["appVersion"];
    ***REMOVED***

    get platform() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get platform' called on an object that is not a valid instance of Navigator.");
      ***REMOVED***

      return esValue[implSymbol]["platform"];
    ***REMOVED***

    get product() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get product' called on an object that is not a valid instance of Navigator.");
      ***REMOVED***

      return esValue[implSymbol]["product"];
    ***REMOVED***

    get productSub() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get productSub' called on an object that is not a valid instance of Navigator.");
      ***REMOVED***

      return esValue[implSymbol]["productSub"];
    ***REMOVED***

    get userAgent() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get userAgent' called on an object that is not a valid instance of Navigator.");
      ***REMOVED***

      return esValue[implSymbol]["userAgent"];
    ***REMOVED***

    get vendor() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get vendor' called on an object that is not a valid instance of Navigator.");
      ***REMOVED***

      return esValue[implSymbol]["vendor"];
    ***REMOVED***

    get vendorSub() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get vendorSub' called on an object that is not a valid instance of Navigator.");
      ***REMOVED***

      return esValue[implSymbol]["vendorSub"];
    ***REMOVED***

    get language() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get language' called on an object that is not a valid instance of Navigator.");
      ***REMOVED***

      return esValue[implSymbol]["language"];
    ***REMOVED***

    get languages() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get languages' called on an object that is not a valid instance of Navigator.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["languages"]);
    ***REMOVED***

    get onLine() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onLine' called on an object that is not a valid instance of Navigator.");
      ***REMOVED***

      return esValue[implSymbol]["onLine"];
    ***REMOVED***

    get cookieEnabled() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get cookieEnabled' called on an object that is not a valid instance of Navigator.");
      ***REMOVED***

      return esValue[implSymbol]["cookieEnabled"];
    ***REMOVED***

    get plugins() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get plugins' called on an object that is not a valid instance of Navigator.");
      ***REMOVED***

      return utils.getSameObject(this, "plugins", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["plugins"]);
      ***REMOVED***);
    ***REMOVED***

    get mimeTypes() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get mimeTypes' called on an object that is not a valid instance of Navigator.");
      ***REMOVED***

      return utils.getSameObject(this, "mimeTypes", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["mimeTypes"]);
      ***REMOVED***);
    ***REMOVED***

    get hardwareConcurrency() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get hardwareConcurrency' called on an object that is not a valid instance of Navigator.");
      ***REMOVED***

      return esValue[implSymbol]["hardwareConcurrency"];
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(Navigator.prototype, {
    javaEnabled: { enumerable: true ***REMOVED***,
    appCodeName: { enumerable: true ***REMOVED***,
    appName: { enumerable: true ***REMOVED***,
    appVersion: { enumerable: true ***REMOVED***,
    platform: { enumerable: true ***REMOVED***,
    product: { enumerable: true ***REMOVED***,
    productSub: { enumerable: true ***REMOVED***,
    userAgent: { enumerable: true ***REMOVED***,
    vendor: { enumerable: true ***REMOVED***,
    vendorSub: { enumerable: true ***REMOVED***,
    language: { enumerable: true ***REMOVED***,
    languages: { enumerable: true ***REMOVED***,
    onLine: { enumerable: true ***REMOVED***,
    cookieEnabled: { enumerable: true ***REMOVED***,
    plugins: { enumerable: true ***REMOVED***,
    mimeTypes: { enumerable: true ***REMOVED***,
    hardwareConcurrency: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "Navigator", configurable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = Navigator;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: Navigator
  ***REMOVED***);
***REMOVED***;

const Impl = require("../navigator/Navigator-impl.js");
