// jsdom/lib/jsdom/living/generated/Node.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const GetRootNodeOptions = require("./GetRootNodeOptions.js");
const ceReactionsPreSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPreSteps;
const ceReactionsPostSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPostSteps;
const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;
const EventTarget = require("./EventTarget.js");

const interfaceName = "Node";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'Node'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["Node"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor Node is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {
  EventTarget._internalSetup(wrapper, globalObject);
***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***

  if (globalObject.EventTarget === undefined) {
    throw new Error("Internal error: attempting to evaluate Node before EventTarget");
  ***REMOVED***
  class Node extends globalObject.EventTarget {
    constructor() {
      throw new TypeError("Illegal constructor");
    ***REMOVED***

    getRootNode() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'getRootNode' called on an object that is not a valid instance of Node.");
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = GetRootNodeOptions.convert(curArg, {
          context: "Failed to execute 'getRootNode' on 'Node': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].getRootNode(...args));
    ***REMOVED***

    hasChildNodes() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'hasChildNodes' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      return esValue[implSymbol].hasChildNodes();
    ***REMOVED***

    normalize() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'normalize' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].normalize();
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    cloneNode() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'cloneNode' called on an object that is not a valid instance of Node.");
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        if (curArg !== undefined) {
          curArg = conversions["boolean"](curArg, { context: "Failed to execute 'cloneNode' on 'Node': parameter 1" ***REMOVED***);
        ***REMOVED*** else {
          curArg = false;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return utils.tryWrapperForImpl(esValue[implSymbol].cloneNode(...args));
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    isEqualNode(otherNode) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'isEqualNode' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'isEqualNode' on 'Node': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        if (curArg === null || curArg === undefined) {
          curArg = null;
        ***REMOVED*** else {
          curArg = exports.convert(curArg, { context: "Failed to execute 'isEqualNode' on 'Node': parameter 1" ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].isEqualNode(...args);
    ***REMOVED***

    isSameNode(otherNode) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'isSameNode' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'isSameNode' on 'Node': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        if (curArg === null || curArg === undefined) {
          curArg = null;
        ***REMOVED*** else {
          curArg = exports.convert(curArg, { context: "Failed to execute 'isSameNode' on 'Node': parameter 1" ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].isSameNode(...args);
    ***REMOVED***

    compareDocumentPosition(other) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'compareDocumentPosition' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'compareDocumentPosition' on 'Node': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = exports.convert(curArg, {
          context: "Failed to execute 'compareDocumentPosition' on 'Node': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].compareDocumentPosition(...args);
    ***REMOVED***

    contains(other) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'contains' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'contains' on 'Node': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        if (curArg === null || curArg === undefined) {
          curArg = null;
        ***REMOVED*** else {
          curArg = exports.convert(curArg, { context: "Failed to execute 'contains' on 'Node': parameter 1" ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].contains(...args);
    ***REMOVED***

    lookupPrefix(namespace) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'lookupPrefix' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'lookupPrefix' on 'Node': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        if (curArg === null || curArg === undefined) {
          curArg = null;
        ***REMOVED*** else {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'lookupPrefix' on 'Node': parameter 1"
          ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].lookupPrefix(...args);
    ***REMOVED***

    lookupNamespaceURI(prefix) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'lookupNamespaceURI' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'lookupNamespaceURI' on 'Node': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        if (curArg === null || curArg === undefined) {
          curArg = null;
        ***REMOVED*** else {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'lookupNamespaceURI' on 'Node': parameter 1"
          ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].lookupNamespaceURI(...args);
    ***REMOVED***

    isDefaultNamespace(namespace) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'isDefaultNamespace' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'isDefaultNamespace' on 'Node': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        if (curArg === null || curArg === undefined) {
          curArg = null;
        ***REMOVED*** else {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'isDefaultNamespace' on 'Node': parameter 1"
          ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].isDefaultNamespace(...args);
    ***REMOVED***

    insertBefore(node, child) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'insertBefore' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      if (arguments.length < 2) {
        throw new TypeError(
          "Failed to execute 'insertBefore' on 'Node': 2 arguments required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = exports.convert(curArg, { context: "Failed to execute 'insertBefore' on 'Node': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        if (curArg === null || curArg === undefined) {
          curArg = null;
        ***REMOVED*** else {
          curArg = exports.convert(curArg, { context: "Failed to execute 'insertBefore' on 'Node': parameter 2" ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return utils.tryWrapperForImpl(esValue[implSymbol].insertBefore(...args));
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    appendChild(node) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'appendChild' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'appendChild' on 'Node': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = exports.convert(curArg, { context: "Failed to execute 'appendChild' on 'Node': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return utils.tryWrapperForImpl(esValue[implSymbol].appendChild(...args));
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    replaceChild(node, child) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'replaceChild' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      if (arguments.length < 2) {
        throw new TypeError(
          "Failed to execute 'replaceChild' on 'Node': 2 arguments required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = exports.convert(curArg, { context: "Failed to execute 'replaceChild' on 'Node': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        curArg = exports.convert(curArg, { context: "Failed to execute 'replaceChild' on 'Node': parameter 2" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return utils.tryWrapperForImpl(esValue[implSymbol].replaceChild(...args));
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    removeChild(child) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'removeChild' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'removeChild' on 'Node': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = exports.convert(curArg, { context: "Failed to execute 'removeChild' on 'Node': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return utils.tryWrapperForImpl(esValue[implSymbol].removeChild(...args));
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get nodeType() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get nodeType' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      return esValue[implSymbol]["nodeType"];
    ***REMOVED***

    get nodeName() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get nodeName' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      return esValue[implSymbol]["nodeName"];
    ***REMOVED***

    get baseURI() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get baseURI' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      return esValue[implSymbol]["baseURI"];
    ***REMOVED***

    get isConnected() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get isConnected' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      return esValue[implSymbol]["isConnected"];
    ***REMOVED***

    get ownerDocument() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get ownerDocument' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["ownerDocument"]);
    ***REMOVED***

    get parentNode() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get parentNode' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["parentNode"]);
    ***REMOVED***

    get parentElement() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get parentElement' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["parentElement"]);
    ***REMOVED***

    get childNodes() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get childNodes' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      return utils.getSameObject(this, "childNodes", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["childNodes"]);
      ***REMOVED***);
    ***REMOVED***

    get firstChild() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get firstChild' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["firstChild"]);
    ***REMOVED***

    get lastChild() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get lastChild' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["lastChild"]);
    ***REMOVED***

    get previousSibling() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get previousSibling' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["previousSibling"]);
    ***REMOVED***

    get nextSibling() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get nextSibling' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["nextSibling"]);
    ***REMOVED***

    get nodeValue() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get nodeValue' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol]["nodeValue"];
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set nodeValue(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set nodeValue' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      if (V === null || V === undefined) {
        V = null;
      ***REMOVED*** else {
        V = conversions["DOMString"](V, {
          context: "Failed to set the 'nodeValue' property on 'Node': The provided value"
        ***REMOVED***);
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol]["nodeValue"] = V;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    get textContent() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get textContent' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol]["textContent"];
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    set textContent(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set textContent' called on an object that is not a valid instance of Node.");
      ***REMOVED***

      if (V === null || V === undefined) {
        V = null;
      ***REMOVED*** else {
        V = conversions["DOMString"](V, {
          context: "Failed to set the 'textContent' property on 'Node': The provided value"
        ***REMOVED***);
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        esValue[implSymbol]["textContent"] = V;
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(Node.prototype, {
    getRootNode: { enumerable: true ***REMOVED***,
    hasChildNodes: { enumerable: true ***REMOVED***,
    normalize: { enumerable: true ***REMOVED***,
    cloneNode: { enumerable: true ***REMOVED***,
    isEqualNode: { enumerable: true ***REMOVED***,
    isSameNode: { enumerable: true ***REMOVED***,
    compareDocumentPosition: { enumerable: true ***REMOVED***,
    contains: { enumerable: true ***REMOVED***,
    lookupPrefix: { enumerable: true ***REMOVED***,
    lookupNamespaceURI: { enumerable: true ***REMOVED***,
    isDefaultNamespace: { enumerable: true ***REMOVED***,
    insertBefore: { enumerable: true ***REMOVED***,
    appendChild: { enumerable: true ***REMOVED***,
    replaceChild: { enumerable: true ***REMOVED***,
    removeChild: { enumerable: true ***REMOVED***,
    nodeType: { enumerable: true ***REMOVED***,
    nodeName: { enumerable: true ***REMOVED***,
    baseURI: { enumerable: true ***REMOVED***,
    isConnected: { enumerable: true ***REMOVED***,
    ownerDocument: { enumerable: true ***REMOVED***,
    parentNode: { enumerable: true ***REMOVED***,
    parentElement: { enumerable: true ***REMOVED***,
    childNodes: { enumerable: true ***REMOVED***,
    firstChild: { enumerable: true ***REMOVED***,
    lastChild: { enumerable: true ***REMOVED***,
    previousSibling: { enumerable: true ***REMOVED***,
    nextSibling: { enumerable: true ***REMOVED***,
    nodeValue: { enumerable: true ***REMOVED***,
    textContent: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "Node", configurable: true ***REMOVED***,
    ELEMENT_NODE: { value: 1, enumerable: true ***REMOVED***,
    ATTRIBUTE_NODE: { value: 2, enumerable: true ***REMOVED***,
    TEXT_NODE: { value: 3, enumerable: true ***REMOVED***,
    CDATA_SECTION_NODE: { value: 4, enumerable: true ***REMOVED***,
    ENTITY_REFERENCE_NODE: { value: 5, enumerable: true ***REMOVED***,
    ENTITY_NODE: { value: 6, enumerable: true ***REMOVED***,
    PROCESSING_INSTRUCTION_NODE: { value: 7, enumerable: true ***REMOVED***,
    COMMENT_NODE: { value: 8, enumerable: true ***REMOVED***,
    DOCUMENT_NODE: { value: 9, enumerable: true ***REMOVED***,
    DOCUMENT_TYPE_NODE: { value: 10, enumerable: true ***REMOVED***,
    DOCUMENT_FRAGMENT_NODE: { value: 11, enumerable: true ***REMOVED***,
    NOTATION_NODE: { value: 12, enumerable: true ***REMOVED***,
    DOCUMENT_POSITION_DISCONNECTED: { value: 0x01, enumerable: true ***REMOVED***,
    DOCUMENT_POSITION_PRECEDING: { value: 0x02, enumerable: true ***REMOVED***,
    DOCUMENT_POSITION_FOLLOWING: { value: 0x04, enumerable: true ***REMOVED***,
    DOCUMENT_POSITION_CONTAINS: { value: 0x08, enumerable: true ***REMOVED***,
    DOCUMENT_POSITION_CONTAINED_BY: { value: 0x10, enumerable: true ***REMOVED***,
    DOCUMENT_POSITION_IMPLEMENTATION_SPECIFIC: { value: 0x20, enumerable: true ***REMOVED***
  ***REMOVED***);
  Object.defineProperties(Node, {
    ELEMENT_NODE: { value: 1, enumerable: true ***REMOVED***,
    ATTRIBUTE_NODE: { value: 2, enumerable: true ***REMOVED***,
    TEXT_NODE: { value: 3, enumerable: true ***REMOVED***,
    CDATA_SECTION_NODE: { value: 4, enumerable: true ***REMOVED***,
    ENTITY_REFERENCE_NODE: { value: 5, enumerable: true ***REMOVED***,
    ENTITY_NODE: { value: 6, enumerable: true ***REMOVED***,
    PROCESSING_INSTRUCTION_NODE: { value: 7, enumerable: true ***REMOVED***,
    COMMENT_NODE: { value: 8, enumerable: true ***REMOVED***,
    DOCUMENT_NODE: { value: 9, enumerable: true ***REMOVED***,
    DOCUMENT_TYPE_NODE: { value: 10, enumerable: true ***REMOVED***,
    DOCUMENT_FRAGMENT_NODE: { value: 11, enumerable: true ***REMOVED***,
    NOTATION_NODE: { value: 12, enumerable: true ***REMOVED***,
    DOCUMENT_POSITION_DISCONNECTED: { value: 0x01, enumerable: true ***REMOVED***,
    DOCUMENT_POSITION_PRECEDING: { value: 0x02, enumerable: true ***REMOVED***,
    DOCUMENT_POSITION_FOLLOWING: { value: 0x04, enumerable: true ***REMOVED***,
    DOCUMENT_POSITION_CONTAINS: { value: 0x08, enumerable: true ***REMOVED***,
    DOCUMENT_POSITION_CONTAINED_BY: { value: 0x10, enumerable: true ***REMOVED***,
    DOCUMENT_POSITION_IMPLEMENTATION_SPECIFIC: { value: 0x20, enumerable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = Node;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: Node
  ***REMOVED***);
***REMOVED***;

const Impl = require("../nodes/Node-impl.js");
