// jsdom/lib/jsdom/living/generated/NodeIterator.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;

const interfaceName = "NodeIterator";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'NodeIterator'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["NodeIterator"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor NodeIterator is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***
  class NodeIterator {
    constructor() {
      throw new TypeError("Illegal constructor");
    ***REMOVED***

    nextNode() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'nextNode' called on an object that is not a valid instance of NodeIterator.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol].nextNode());
    ***REMOVED***

    previousNode() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'previousNode' called on an object that is not a valid instance of NodeIterator.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol].previousNode());
    ***REMOVED***

    detach() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'detach' called on an object that is not a valid instance of NodeIterator.");
      ***REMOVED***

      return esValue[implSymbol].detach();
    ***REMOVED***

    get root() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get root' called on an object that is not a valid instance of NodeIterator.");
      ***REMOVED***

      return utils.getSameObject(this, "root", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["root"]);
      ***REMOVED***);
    ***REMOVED***

    get referenceNode() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get referenceNode' called on an object that is not a valid instance of NodeIterator.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["referenceNode"]);
    ***REMOVED***

    get pointerBeforeReferenceNode() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get pointerBeforeReferenceNode' called on an object that is not a valid instance of NodeIterator."
        );
      ***REMOVED***

      return esValue[implSymbol]["pointerBeforeReferenceNode"];
    ***REMOVED***

    get whatToShow() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get whatToShow' called on an object that is not a valid instance of NodeIterator.");
      ***REMOVED***

      return esValue[implSymbol]["whatToShow"];
    ***REMOVED***

    get filter() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get filter' called on an object that is not a valid instance of NodeIterator.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["filter"]);
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(NodeIterator.prototype, {
    nextNode: { enumerable: true ***REMOVED***,
    previousNode: { enumerable: true ***REMOVED***,
    detach: { enumerable: true ***REMOVED***,
    root: { enumerable: true ***REMOVED***,
    referenceNode: { enumerable: true ***REMOVED***,
    pointerBeforeReferenceNode: { enumerable: true ***REMOVED***,
    whatToShow: { enumerable: true ***REMOVED***,
    filter: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "NodeIterator", configurable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = NodeIterator;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: NodeIterator
  ***REMOVED***);
***REMOVED***;

const Impl = require("../traversal/NodeIterator-impl.js");
