// jsdom/lib/jsdom/living/generated/NodeFilter.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

exports.convert = function convert(value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) {
  if (!utils.isObject(value)) {
    throw new TypeError(`${context***REMOVED*** is not an object.`);
  ***REMOVED***

  function callTheUserObjectsOperation(node) {
    let thisArg = utils.tryWrapperForImpl(this);
    let O = value;
    let X = O;

    if (typeof O !== "function") {
      X = O["acceptNode"];
      if (typeof X !== "function") {
        throw new TypeError(`${context***REMOVED*** does not correctly implement NodeFilter.`);
      ***REMOVED***
      thisArg = O;
    ***REMOVED***

    node = utils.tryWrapperForImpl(node);

    let callResult = Reflect.apply(X, thisArg, [node]);

    callResult = conversions["unsigned short"](callResult, { context: context ***REMOVED***);

    return callResult;
  ***REMOVED***

  callTheUserObjectsOperation[utils.wrapperSymbol] = value;
  callTheUserObjectsOperation.objectReference = value;

  return callTheUserObjectsOperation;
***REMOVED***;

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***

  const NodeFilter = () => {
    throw new TypeError("Illegal invocation");
  ***REMOVED***;

  Object.defineProperties(NodeFilter, {
    FILTER_ACCEPT: { value: 1, enumerable: true ***REMOVED***,
    FILTER_REJECT: { value: 2, enumerable: true ***REMOVED***,
    FILTER_SKIP: { value: 3, enumerable: true ***REMOVED***,
    SHOW_ALL: { value: 0xffffffff, enumerable: true ***REMOVED***,
    SHOW_ELEMENT: { value: 0x1, enumerable: true ***REMOVED***,
    SHOW_ATTRIBUTE: { value: 0x2, enumerable: true ***REMOVED***,
    SHOW_TEXT: { value: 0x4, enumerable: true ***REMOVED***,
    SHOW_CDATA_SECTION: { value: 0x8, enumerable: true ***REMOVED***,
    SHOW_ENTITY_REFERENCE: { value: 0x10, enumerable: true ***REMOVED***,
    SHOW_ENTITY: { value: 0x20, enumerable: true ***REMOVED***,
    SHOW_PROCESSING_INSTRUCTION: { value: 0x40, enumerable: true ***REMOVED***,
    SHOW_COMMENT: { value: 0x80, enumerable: true ***REMOVED***,
    SHOW_DOCUMENT: { value: 0x100, enumerable: true ***REMOVED***,
    SHOW_DOCUMENT_TYPE: { value: 0x200, enumerable: true ***REMOVED***,
    SHOW_DOCUMENT_FRAGMENT: { value: 0x400, enumerable: true ***REMOVED***,
    SHOW_NOTATION: { value: 0x800, enumerable: true ***REMOVED***
  ***REMOVED***);

  Object.defineProperty(globalObject, "NodeFilter", {
    configurable: true,
    writable: true,
    value: NodeFilter
  ***REMOVED***);
***REMOVED***;
