// jsdom/lib/jsdom/living/generated/Selection.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const Range = require("./Range.js");
const Node = require("./Node.js");
const ceReactionsPreSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPreSteps;
const ceReactionsPostSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPostSteps;
const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;

const interfaceName = "Selection";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'Selection'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["Selection"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor Selection is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***
  class Selection {
    constructor() {
      throw new TypeError("Illegal constructor");
    ***REMOVED***

    getRangeAt(index) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'getRangeAt' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'getRangeAt' on 'Selection': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["unsigned long"](curArg, {
          context: "Failed to execute 'getRangeAt' on 'Selection': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return utils.tryWrapperForImpl(esValue[implSymbol].getRangeAt(...args));
    ***REMOVED***

    addRange(range) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'addRange' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'addRange' on 'Selection': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = Range.convert(curArg, { context: "Failed to execute 'addRange' on 'Selection': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].addRange(...args);
    ***REMOVED***

    removeRange(range) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'removeRange' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'removeRange' on 'Selection': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = Range.convert(curArg, { context: "Failed to execute 'removeRange' on 'Selection': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].removeRange(...args);
    ***REMOVED***

    removeAllRanges() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'removeAllRanges' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      return esValue[implSymbol].removeAllRanges();
    ***REMOVED***

    empty() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'empty' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      return esValue[implSymbol].empty();
    ***REMOVED***

    collapse(node) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'collapse' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'collapse' on 'Selection': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        if (curArg === null || curArg === undefined) {
          curArg = null;
        ***REMOVED*** else {
          curArg = Node.convert(curArg, { context: "Failed to execute 'collapse' on 'Selection': parameter 1" ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        if (curArg !== undefined) {
          curArg = conversions["unsigned long"](curArg, {
            context: "Failed to execute 'collapse' on 'Selection': parameter 2"
          ***REMOVED***);
        ***REMOVED*** else {
          curArg = 0;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].collapse(...args);
    ***REMOVED***

    setPosition(node) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'setPosition' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'setPosition' on 'Selection': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        if (curArg === null || curArg === undefined) {
          curArg = null;
        ***REMOVED*** else {
          curArg = Node.convert(curArg, { context: "Failed to execute 'setPosition' on 'Selection': parameter 1" ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        if (curArg !== undefined) {
          curArg = conversions["unsigned long"](curArg, {
            context: "Failed to execute 'setPosition' on 'Selection': parameter 2"
          ***REMOVED***);
        ***REMOVED*** else {
          curArg = 0;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].setPosition(...args);
    ***REMOVED***

    collapseToStart() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'collapseToStart' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      return esValue[implSymbol].collapseToStart();
    ***REMOVED***

    collapseToEnd() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'collapseToEnd' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      return esValue[implSymbol].collapseToEnd();
    ***REMOVED***

    extend(node) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'extend' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'extend' on 'Selection': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = Node.convert(curArg, { context: "Failed to execute 'extend' on 'Selection': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        if (curArg !== undefined) {
          curArg = conversions["unsigned long"](curArg, {
            context: "Failed to execute 'extend' on 'Selection': parameter 2"
          ***REMOVED***);
        ***REMOVED*** else {
          curArg = 0;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].extend(...args);
    ***REMOVED***

    setBaseAndExtent(anchorNode, anchorOffset, focusNode, focusOffset) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'setBaseAndExtent' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      if (arguments.length < 4) {
        throw new TypeError(
          "Failed to execute 'setBaseAndExtent' on 'Selection': 4 arguments required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = Node.convert(curArg, { context: "Failed to execute 'setBaseAndExtent' on 'Selection': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        curArg = conversions["unsigned long"](curArg, {
          context: "Failed to execute 'setBaseAndExtent' on 'Selection': parameter 2"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[2];
        curArg = Node.convert(curArg, { context: "Failed to execute 'setBaseAndExtent' on 'Selection': parameter 3" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[3];
        curArg = conversions["unsigned long"](curArg, {
          context: "Failed to execute 'setBaseAndExtent' on 'Selection': parameter 4"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].setBaseAndExtent(...args);
    ***REMOVED***

    selectAllChildren(node) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'selectAllChildren' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'selectAllChildren' on 'Selection': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = Node.convert(curArg, { context: "Failed to execute 'selectAllChildren' on 'Selection': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].selectAllChildren(...args);
    ***REMOVED***

    deleteFromDocument() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'deleteFromDocument' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].deleteFromDocument();
      ***REMOVED*** finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      ***REMOVED***
    ***REMOVED***

    containsNode(node) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'containsNode' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'containsNode' on 'Selection': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = Node.convert(curArg, { context: "Failed to execute 'containsNode' on 'Selection': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        if (curArg !== undefined) {
          curArg = conversions["boolean"](curArg, {
            context: "Failed to execute 'containsNode' on 'Selection': parameter 2"
          ***REMOVED***);
        ***REMOVED*** else {
          curArg = false;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].containsNode(...args);
    ***REMOVED***

    toString() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'toString' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      return esValue[implSymbol].toString();
    ***REMOVED***

    get anchorNode() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get anchorNode' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["anchorNode"]);
    ***REMOVED***

    get anchorOffset() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get anchorOffset' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      return esValue[implSymbol]["anchorOffset"];
    ***REMOVED***

    get focusNode() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get focusNode' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["focusNode"]);
    ***REMOVED***

    get focusOffset() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get focusOffset' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      return esValue[implSymbol]["focusOffset"];
    ***REMOVED***

    get isCollapsed() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get isCollapsed' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      return esValue[implSymbol]["isCollapsed"];
    ***REMOVED***

    get rangeCount() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get rangeCount' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      return esValue[implSymbol]["rangeCount"];
    ***REMOVED***

    get type() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get type' called on an object that is not a valid instance of Selection.");
      ***REMOVED***

      return esValue[implSymbol]["type"];
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(Selection.prototype, {
    getRangeAt: { enumerable: true ***REMOVED***,
    addRange: { enumerable: true ***REMOVED***,
    removeRange: { enumerable: true ***REMOVED***,
    removeAllRanges: { enumerable: true ***REMOVED***,
    empty: { enumerable: true ***REMOVED***,
    collapse: { enumerable: true ***REMOVED***,
    setPosition: { enumerable: true ***REMOVED***,
    collapseToStart: { enumerable: true ***REMOVED***,
    collapseToEnd: { enumerable: true ***REMOVED***,
    extend: { enumerable: true ***REMOVED***,
    setBaseAndExtent: { enumerable: true ***REMOVED***,
    selectAllChildren: { enumerable: true ***REMOVED***,
    deleteFromDocument: { enumerable: true ***REMOVED***,
    containsNode: { enumerable: true ***REMOVED***,
    toString: { enumerable: true ***REMOVED***,
    anchorNode: { enumerable: true ***REMOVED***,
    anchorOffset: { enumerable: true ***REMOVED***,
    focusNode: { enumerable: true ***REMOVED***,
    focusOffset: { enumerable: true ***REMOVED***,
    isCollapsed: { enumerable: true ***REMOVED***,
    rangeCount: { enumerable: true ***REMOVED***,
    type: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "Selection", configurable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = Selection;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: Selection
  ***REMOVED***);
***REMOVED***;

const Impl = require("../selection/Selection-impl.js");
