// jsdom/lib/jsdom/living/generated/Storage.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;

const interfaceName = "Storage";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'Storage'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["Storage"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor Storage is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper = new Proxy(wrapper, proxyHandler);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  let wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper = new Proxy(wrapper, proxyHandler);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***
  class Storage {
    constructor() {
      throw new TypeError("Illegal constructor");
    ***REMOVED***

    key(index) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'key' called on an object that is not a valid instance of Storage.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'key' on 'Storage': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["unsigned long"](curArg, { context: "Failed to execute 'key' on 'Storage': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].key(...args);
    ***REMOVED***

    getItem(key) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'getItem' called on an object that is not a valid instance of Storage.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'getItem' on 'Storage': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, { context: "Failed to execute 'getItem' on 'Storage': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].getItem(...args);
    ***REMOVED***

    setItem(key, value) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'setItem' called on an object that is not a valid instance of Storage.");
      ***REMOVED***

      if (arguments.length < 2) {
        throw new TypeError(
          "Failed to execute 'setItem' on 'Storage': 2 arguments required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, { context: "Failed to execute 'setItem' on 'Storage': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        curArg = conversions["DOMString"](curArg, { context: "Failed to execute 'setItem' on 'Storage': parameter 2" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].setItem(...args);
    ***REMOVED***

    removeItem(key) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'removeItem' called on an object that is not a valid instance of Storage.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'removeItem' on 'Storage': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'removeItem' on 'Storage': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].removeItem(...args);
    ***REMOVED***

    clear() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'clear' called on an object that is not a valid instance of Storage.");
      ***REMOVED***

      return esValue[implSymbol].clear();
    ***REMOVED***

    get length() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get length' called on an object that is not a valid instance of Storage.");
      ***REMOVED***

      return esValue[implSymbol]["length"];
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(Storage.prototype, {
    key: { enumerable: true ***REMOVED***,
    getItem: { enumerable: true ***REMOVED***,
    setItem: { enumerable: true ***REMOVED***,
    removeItem: { enumerable: true ***REMOVED***,
    clear: { enumerable: true ***REMOVED***,
    length: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "Storage", configurable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = Storage;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: Storage
  ***REMOVED***);
***REMOVED***;

const proxyHandler = {
  get(target, P, receiver) {
    if (typeof P === "symbol") {
      return Reflect.get(target, P, receiver);
    ***REMOVED***
    const desc = this.getOwnPropertyDescriptor(target, P);
    if (desc === undefined) {
      const parent = Object.getPrototypeOf(target);
      if (parent === null) {
        return undefined;
      ***REMOVED***
      return Reflect.get(target, P, receiver);
    ***REMOVED***
    if (!desc.get && !desc.set) {
      return desc.value;
    ***REMOVED***
    const getter = desc.get;
    if (getter === undefined) {
      return undefined;
    ***REMOVED***
    return Reflect.apply(getter, receiver, []);
  ***REMOVED***,

  has(target, P) {
    if (typeof P === "symbol") {
      return Reflect.has(target, P);
    ***REMOVED***
    const desc = this.getOwnPropertyDescriptor(target, P);
    if (desc !== undefined) {
      return true;
    ***REMOVED***
    const parent = Object.getPrototypeOf(target);
    if (parent !== null) {
      return Reflect.has(parent, P);
    ***REMOVED***
    return false;
  ***REMOVED***,

  ownKeys(target) {
    const keys = new Set();

    for (const key of target[implSymbol][utils.supportedPropertyNames]) {
      if (!(key in target)) {
        keys.add(`${key***REMOVED***`);
      ***REMOVED***
    ***REMOVED***

    for (const key of Reflect.ownKeys(target)) {
      keys.add(key);
    ***REMOVED***
    return [...keys];
  ***REMOVED***,

  getOwnPropertyDescriptor(target, P) {
    if (typeof P === "symbol") {
      return Reflect.getOwnPropertyDescriptor(target, P);
    ***REMOVED***
    let ignoreNamedProps = false;

    const namedValue = target[implSymbol].getItem(P);

    if (namedValue !== null && !(P in target) && !ignoreNamedProps) {
      return {
        writable: true,
        enumerable: true,
        configurable: true,
        value: utils.tryWrapperForImpl(namedValue)
      ***REMOVED***;
    ***REMOVED***

    return Reflect.getOwnPropertyDescriptor(target, P);
  ***REMOVED***,

  set(target, P, V, receiver) {
    if (typeof P === "symbol") {
      return Reflect.set(target, P, V, receiver);
    ***REMOVED***
    // The `receiver` argument refers to the Proxy exotic object or an object
    // that inherits from it, whereas `target` refers to the Proxy target:
    if (target[implSymbol][utils.wrapperSymbol] === receiver) {
      if (typeof P === "string") {
        let namedValue = V;

        namedValue = conversions["DOMString"](namedValue, {
          context: "Failed to set the '" + P + "' property on 'Storage': The provided value"
        ***REMOVED***);

        target[implSymbol].setItem(P, namedValue);

        return true;
      ***REMOVED***
    ***REMOVED***
    let ownDesc;

    if (ownDesc === undefined) {
      ownDesc = Reflect.getOwnPropertyDescriptor(target, P);
    ***REMOVED***
    if (ownDesc === undefined) {
      const parent = Reflect.getPrototypeOf(target);
      if (parent !== null) {
        return Reflect.set(parent, P, V, receiver);
      ***REMOVED***
      ownDesc = { writable: true, enumerable: true, configurable: true, value: undefined ***REMOVED***;
    ***REMOVED***
    if (!ownDesc.writable) {
      return false;
    ***REMOVED***
    if (!utils.isObject(receiver)) {
      return false;
    ***REMOVED***
    const existingDesc = Reflect.getOwnPropertyDescriptor(receiver, P);
    let valueDesc;
    if (existingDesc !== undefined) {
      if (existingDesc.get || existingDesc.set) {
        return false;
      ***REMOVED***
      if (!existingDesc.writable) {
        return false;
      ***REMOVED***
      valueDesc = { value: V ***REMOVED***;
    ***REMOVED*** else {
      valueDesc = { writable: true, enumerable: true, configurable: true, value: V ***REMOVED***;
    ***REMOVED***
    return Reflect.defineProperty(receiver, P, valueDesc);
  ***REMOVED***,

  defineProperty(target, P, desc) {
    if (typeof P === "symbol") {
      return Reflect.defineProperty(target, P, desc);
    ***REMOVED***
    if (!utils.hasOwn(target, P)) {
      if (desc.get || desc.set) {
        return false;
      ***REMOVED***

      let namedValue = desc.value;

      namedValue = conversions["DOMString"](namedValue, {
        context: "Failed to set the '" + P + "' property on 'Storage': The provided value"
      ***REMOVED***);

      target[implSymbol].setItem(P, namedValue);

      return true;
    ***REMOVED***
    return Reflect.defineProperty(target, P, desc);
  ***REMOVED***,

  deleteProperty(target, P) {
    if (typeof P === "symbol") {
      return Reflect.deleteProperty(target, P);
    ***REMOVED***

    if (target[implSymbol].getItem(P) !== null && !(P in target)) {
      target[implSymbol].removeItem(P);
      return true;
    ***REMOVED***

    return Reflect.deleteProperty(target, P);
  ***REMOVED***,

  preventExtensions() {
    return false;
  ***REMOVED***
***REMOVED***;

const Impl = require("../webstorage/Storage-impl.js");
