// jsdom/lib/jsdom/living/generated/StorageEvent.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const StorageEventInit = require("./StorageEventInit.js");
const Storage = require("./Storage.js");
const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;
const Event = require("./Event.js");

const interfaceName = "StorageEvent";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'StorageEvent'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["StorageEvent"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor StorageEvent is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {
  Event._internalSetup(wrapper, globalObject);
***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***

  if (globalObject.Event === undefined) {
    throw new Error("Internal error: attempting to evaluate StorageEvent before Event");
  ***REMOVED***
  class StorageEvent extends globalObject.Event {
    constructor(type) {
      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to construct 'StorageEvent': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, { context: "Failed to construct 'StorageEvent': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        curArg = StorageEventInit.convert(curArg, { context: "Failed to construct 'StorageEvent': parameter 2" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return exports.setup(Object.create(new.target.prototype), globalObject, args);
    ***REMOVED***

    initStorageEvent(type) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'initStorageEvent' called on an object that is not a valid instance of StorageEvent.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'initStorageEvent' on 'StorageEvent': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'initStorageEvent' on 'StorageEvent': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        if (curArg !== undefined) {
          curArg = conversions["boolean"](curArg, {
            context: "Failed to execute 'initStorageEvent' on 'StorageEvent': parameter 2"
          ***REMOVED***);
        ***REMOVED*** else {
          curArg = false;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[2];
        if (curArg !== undefined) {
          curArg = conversions["boolean"](curArg, {
            context: "Failed to execute 'initStorageEvent' on 'StorageEvent': parameter 3"
          ***REMOVED***);
        ***REMOVED*** else {
          curArg = false;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[3];
        if (curArg !== undefined) {
          if (curArg === null || curArg === undefined) {
            curArg = null;
          ***REMOVED*** else {
            curArg = conversions["DOMString"](curArg, {
              context: "Failed to execute 'initStorageEvent' on 'StorageEvent': parameter 4"
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED*** else {
          curArg = null;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[4];
        if (curArg !== undefined) {
          if (curArg === null || curArg === undefined) {
            curArg = null;
          ***REMOVED*** else {
            curArg = conversions["DOMString"](curArg, {
              context: "Failed to execute 'initStorageEvent' on 'StorageEvent': parameter 5"
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED*** else {
          curArg = null;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[5];
        if (curArg !== undefined) {
          if (curArg === null || curArg === undefined) {
            curArg = null;
          ***REMOVED*** else {
            curArg = conversions["DOMString"](curArg, {
              context: "Failed to execute 'initStorageEvent' on 'StorageEvent': parameter 6"
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED*** else {
          curArg = null;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[6];
        if (curArg !== undefined) {
          curArg = conversions["USVString"](curArg, {
            context: "Failed to execute 'initStorageEvent' on 'StorageEvent': parameter 7"
          ***REMOVED***);
        ***REMOVED*** else {
          curArg = "";
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[7];
        if (curArg !== undefined) {
          if (curArg === null || curArg === undefined) {
            curArg = null;
          ***REMOVED*** else {
            curArg = Storage.convert(curArg, {
              context: "Failed to execute 'initStorageEvent' on 'StorageEvent': parameter 8"
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED*** else {
          curArg = null;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].initStorageEvent(...args);
    ***REMOVED***

    get key() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get key' called on an object that is not a valid instance of StorageEvent.");
      ***REMOVED***

      return esValue[implSymbol]["key"];
    ***REMOVED***

    get oldValue() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get oldValue' called on an object that is not a valid instance of StorageEvent.");
      ***REMOVED***

      return esValue[implSymbol]["oldValue"];
    ***REMOVED***

    get newValue() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get newValue' called on an object that is not a valid instance of StorageEvent.");
      ***REMOVED***

      return esValue[implSymbol]["newValue"];
    ***REMOVED***

    get url() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get url' called on an object that is not a valid instance of StorageEvent.");
      ***REMOVED***

      return esValue[implSymbol]["url"];
    ***REMOVED***

    get storageArea() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get storageArea' called on an object that is not a valid instance of StorageEvent.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["storageArea"]);
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(StorageEvent.prototype, {
    initStorageEvent: { enumerable: true ***REMOVED***,
    key: { enumerable: true ***REMOVED***,
    oldValue: { enumerable: true ***REMOVED***,
    newValue: { enumerable: true ***REMOVED***,
    url: { enumerable: true ***REMOVED***,
    storageArea: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "StorageEvent", configurable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = StorageEvent;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: StorageEvent
  ***REMOVED***);
***REMOVED***;

const Impl = require("../events/StorageEvent-impl.js");
