// jsdom/lib/jsdom/living/generated/TreeWalker.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const Node = require("./Node.js");
const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;

const interfaceName = "TreeWalker";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'TreeWalker'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["TreeWalker"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor TreeWalker is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***
  class TreeWalker {
    constructor() {
      throw new TypeError("Illegal constructor");
    ***REMOVED***

    parentNode() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'parentNode' called on an object that is not a valid instance of TreeWalker.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol].parentNode());
    ***REMOVED***

    firstChild() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'firstChild' called on an object that is not a valid instance of TreeWalker.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol].firstChild());
    ***REMOVED***

    lastChild() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'lastChild' called on an object that is not a valid instance of TreeWalker.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol].lastChild());
    ***REMOVED***

    previousSibling() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'previousSibling' called on an object that is not a valid instance of TreeWalker.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol].previousSibling());
    ***REMOVED***

    nextSibling() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'nextSibling' called on an object that is not a valid instance of TreeWalker.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol].nextSibling());
    ***REMOVED***

    previousNode() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'previousNode' called on an object that is not a valid instance of TreeWalker.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol].previousNode());
    ***REMOVED***

    nextNode() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'nextNode' called on an object that is not a valid instance of TreeWalker.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol].nextNode());
    ***REMOVED***

    get root() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get root' called on an object that is not a valid instance of TreeWalker.");
      ***REMOVED***

      return utils.getSameObject(this, "root", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["root"]);
      ***REMOVED***);
    ***REMOVED***

    get whatToShow() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get whatToShow' called on an object that is not a valid instance of TreeWalker.");
      ***REMOVED***

      return esValue[implSymbol]["whatToShow"];
    ***REMOVED***

    get filter() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get filter' called on an object that is not a valid instance of TreeWalker.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["filter"]);
    ***REMOVED***

    get currentNode() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get currentNode' called on an object that is not a valid instance of TreeWalker.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["currentNode"]);
    ***REMOVED***

    set currentNode(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set currentNode' called on an object that is not a valid instance of TreeWalker.");
      ***REMOVED***

      V = Node.convert(V, { context: "Failed to set the 'currentNode' property on 'TreeWalker': The provided value" ***REMOVED***);

      esValue[implSymbol]["currentNode"] = V;
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(TreeWalker.prototype, {
    parentNode: { enumerable: true ***REMOVED***,
    firstChild: { enumerable: true ***REMOVED***,
    lastChild: { enumerable: true ***REMOVED***,
    previousSibling: { enumerable: true ***REMOVED***,
    nextSibling: { enumerable: true ***REMOVED***,
    previousNode: { enumerable: true ***REMOVED***,
    nextNode: { enumerable: true ***REMOVED***,
    root: { enumerable: true ***REMOVED***,
    whatToShow: { enumerable: true ***REMOVED***,
    filter: { enumerable: true ***REMOVED***,
    currentNode: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "TreeWalker", configurable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = TreeWalker;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: TreeWalker
  ***REMOVED***);
***REMOVED***;

const Impl = require("../traversal/TreeWalker-impl.js");
