// jsdom/lib/jsdom/living/generated/ValidityState.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;

const interfaceName = "ValidityState";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'ValidityState'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["ValidityState"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor ValidityState is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***
  class ValidityState {
    constructor() {
      throw new TypeError("Illegal constructor");
    ***REMOVED***

    get valueMissing() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get valueMissing' called on an object that is not a valid instance of ValidityState.");
      ***REMOVED***

      return esValue[implSymbol]["valueMissing"];
    ***REMOVED***

    get typeMismatch() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get typeMismatch' called on an object that is not a valid instance of ValidityState.");
      ***REMOVED***

      return esValue[implSymbol]["typeMismatch"];
    ***REMOVED***

    get patternMismatch() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get patternMismatch' called on an object that is not a valid instance of ValidityState.");
      ***REMOVED***

      return esValue[implSymbol]["patternMismatch"];
    ***REMOVED***

    get tooLong() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get tooLong' called on an object that is not a valid instance of ValidityState.");
      ***REMOVED***

      return esValue[implSymbol]["tooLong"];
    ***REMOVED***

    get tooShort() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get tooShort' called on an object that is not a valid instance of ValidityState.");
      ***REMOVED***

      return esValue[implSymbol]["tooShort"];
    ***REMOVED***

    get rangeUnderflow() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get rangeUnderflow' called on an object that is not a valid instance of ValidityState.");
      ***REMOVED***

      return esValue[implSymbol]["rangeUnderflow"];
    ***REMOVED***

    get rangeOverflow() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get rangeOverflow' called on an object that is not a valid instance of ValidityState.");
      ***REMOVED***

      return esValue[implSymbol]["rangeOverflow"];
    ***REMOVED***

    get stepMismatch() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get stepMismatch' called on an object that is not a valid instance of ValidityState.");
      ***REMOVED***

      return esValue[implSymbol]["stepMismatch"];
    ***REMOVED***

    get badInput() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get badInput' called on an object that is not a valid instance of ValidityState.");
      ***REMOVED***

      return esValue[implSymbol]["badInput"];
    ***REMOVED***

    get customError() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get customError' called on an object that is not a valid instance of ValidityState.");
      ***REMOVED***

      return esValue[implSymbol]["customError"];
    ***REMOVED***

    get valid() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get valid' called on an object that is not a valid instance of ValidityState.");
      ***REMOVED***

      return esValue[implSymbol]["valid"];
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(ValidityState.prototype, {
    valueMissing: { enumerable: true ***REMOVED***,
    typeMismatch: { enumerable: true ***REMOVED***,
    patternMismatch: { enumerable: true ***REMOVED***,
    tooLong: { enumerable: true ***REMOVED***,
    tooShort: { enumerable: true ***REMOVED***,
    rangeUnderflow: { enumerable: true ***REMOVED***,
    rangeOverflow: { enumerable: true ***REMOVED***,
    stepMismatch: { enumerable: true ***REMOVED***,
    badInput: { enumerable: true ***REMOVED***,
    customError: { enumerable: true ***REMOVED***,
    valid: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "ValidityState", configurable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = ValidityState;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: ValidityState
  ***REMOVED***);
***REMOVED***;

const Impl = require("../constraint-validation/ValidityState-impl.js");
