// jsdom/lib/jsdom/living/generated/WebSocket.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const Blob = require("./Blob.js");
const EventHandlerNonNull = require("./EventHandlerNonNull.js");
const BinaryType = require("./BinaryType.js");
const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;
const EventTarget = require("./EventTarget.js");

const interfaceName = "WebSocket";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'WebSocket'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["WebSocket"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor WebSocket is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {
  EventTarget._internalSetup(wrapper, globalObject);
***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window", "Worker"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***

  if (globalObject.EventTarget === undefined) {
    throw new Error("Internal error: attempting to evaluate WebSocket before EventTarget");
  ***REMOVED***
  class WebSocket extends globalObject.EventTarget {
    constructor(url) {
      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to construct 'WebSocket': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["USVString"](curArg, { context: "Failed to construct 'WebSocket': parameter 1" ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        if (curArg !== undefined) {
          if (utils.isObject(curArg)) {
            if (curArg[Symbol.iterator] !== undefined) {
              if (!utils.isObject(curArg)) {
                throw new TypeError(
                  "Failed to construct 'WebSocket': parameter 2" + " sequence" + " is not an iterable object."
                );
              ***REMOVED*** else {
                const V = [];
                const tmp = curArg;
                for (let nextItem of tmp) {
                  nextItem = conversions["DOMString"](nextItem, {
                    context: "Failed to construct 'WebSocket': parameter 2" + " sequence" + "'s element"
                  ***REMOVED***);

                  V.push(nextItem);
                ***REMOVED***
                curArg = V;
              ***REMOVED***
            ***REMOVED*** else {
            ***REMOVED***
          ***REMOVED*** else {
            curArg = conversions["DOMString"](curArg, { context: "Failed to construct 'WebSocket': parameter 2" ***REMOVED***);
          ***REMOVED***
        ***REMOVED*** else {
          curArg = [];
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return exports.setup(Object.create(new.target.prototype), globalObject, args);
    ***REMOVED***

    close() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'close' called on an object that is not a valid instance of WebSocket.");
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        if (curArg !== undefined) {
          curArg = conversions["unsigned short"](curArg, {
            context: "Failed to execute 'close' on 'WebSocket': parameter 1",
            clamp: true
          ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        if (curArg !== undefined) {
          curArg = conversions["USVString"](curArg, {
            context: "Failed to execute 'close' on 'WebSocket': parameter 2"
          ***REMOVED***);
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].close(...args);
    ***REMOVED***

    send(data) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'send' called on an object that is not a valid instance of WebSocket.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'send' on 'WebSocket': 1 argument required, but only " + arguments.length + " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        if (Blob.is(curArg)) {
          {
            let curArg = arguments[0];
            curArg = Blob.convert(curArg, { context: "Failed to execute 'send' on 'WebSocket': parameter 1" ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
        ***REMOVED*** else if (utils.isArrayBuffer(curArg)) {
          {
            let curArg = arguments[0];
            curArg = conversions["ArrayBuffer"](curArg, {
              context: "Failed to execute 'send' on 'WebSocket': parameter 1"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
        ***REMOVED*** else if (ArrayBuffer.isView(curArg)) {
          {
            let curArg = arguments[0];
            if (ArrayBuffer.isView(curArg)) {
            ***REMOVED*** else {
              throw new TypeError(
                "Failed to execute 'send' on 'WebSocket': parameter 1" + " is not of any supported type."
              );
            ***REMOVED***
            args.push(curArg);
          ***REMOVED***
        ***REMOVED*** else {
          {
            let curArg = arguments[0];
            curArg = conversions["USVString"](curArg, {
              context: "Failed to execute 'send' on 'WebSocket': parameter 1"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
      return esValue[implSymbol].send(...args);
    ***REMOVED***

    get url() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get url' called on an object that is not a valid instance of WebSocket.");
      ***REMOVED***

      return esValue[implSymbol]["url"];
    ***REMOVED***

    get readyState() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get readyState' called on an object that is not a valid instance of WebSocket.");
      ***REMOVED***

      return esValue[implSymbol]["readyState"];
    ***REMOVED***

    get bufferedAmount() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get bufferedAmount' called on an object that is not a valid instance of WebSocket.");
      ***REMOVED***

      return esValue[implSymbol]["bufferedAmount"];
    ***REMOVED***

    get onopen() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onopen' called on an object that is not a valid instance of WebSocket.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onopen"]);
    ***REMOVED***

    set onopen(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onopen' called on an object that is not a valid instance of WebSocket.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onopen' property on 'WebSocket': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onopen"] = V;
    ***REMOVED***

    get onerror() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onerror' called on an object that is not a valid instance of WebSocket.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onerror"]);
    ***REMOVED***

    set onerror(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onerror' called on an object that is not a valid instance of WebSocket.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onerror' property on 'WebSocket': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onerror"] = V;
    ***REMOVED***

    get onclose() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onclose' called on an object that is not a valid instance of WebSocket.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onclose"]);
    ***REMOVED***

    set onclose(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onclose' called on an object that is not a valid instance of WebSocket.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onclose' property on 'WebSocket': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onclose"] = V;
    ***REMOVED***

    get extensions() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get extensions' called on an object that is not a valid instance of WebSocket.");
      ***REMOVED***

      return esValue[implSymbol]["extensions"];
    ***REMOVED***

    get protocol() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get protocol' called on an object that is not a valid instance of WebSocket.");
      ***REMOVED***

      return esValue[implSymbol]["protocol"];
    ***REMOVED***

    get onmessage() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get onmessage' called on an object that is not a valid instance of WebSocket.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onmessage"]);
    ***REMOVED***

    set onmessage(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set onmessage' called on an object that is not a valid instance of WebSocket.");
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onmessage' property on 'WebSocket': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onmessage"] = V;
    ***REMOVED***

    get binaryType() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get binaryType' called on an object that is not a valid instance of WebSocket.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["binaryType"]);
    ***REMOVED***

    set binaryType(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set binaryType' called on an object that is not a valid instance of WebSocket.");
      ***REMOVED***

      V = `${V***REMOVED***`;
      if (!BinaryType.enumerationValues.has(V)) {
        return;
      ***REMOVED***

      esValue[implSymbol]["binaryType"] = V;
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(WebSocket.prototype, {
    close: { enumerable: true ***REMOVED***,
    send: { enumerable: true ***REMOVED***,
    url: { enumerable: true ***REMOVED***,
    readyState: { enumerable: true ***REMOVED***,
    bufferedAmount: { enumerable: true ***REMOVED***,
    onopen: { enumerable: true ***REMOVED***,
    onerror: { enumerable: true ***REMOVED***,
    onclose: { enumerable: true ***REMOVED***,
    extensions: { enumerable: true ***REMOVED***,
    protocol: { enumerable: true ***REMOVED***,
    onmessage: { enumerable: true ***REMOVED***,
    binaryType: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "WebSocket", configurable: true ***REMOVED***,
    CONNECTING: { value: 0, enumerable: true ***REMOVED***,
    OPEN: { value: 1, enumerable: true ***REMOVED***,
    CLOSING: { value: 2, enumerable: true ***REMOVED***,
    CLOSED: { value: 3, enumerable: true ***REMOVED***
  ***REMOVED***);
  Object.defineProperties(WebSocket, {
    CONNECTING: { value: 0, enumerable: true ***REMOVED***,
    OPEN: { value: 1, enumerable: true ***REMOVED***,
    CLOSING: { value: 2, enumerable: true ***REMOVED***,
    CLOSED: { value: 3, enumerable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = WebSocket;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: WebSocket
  ***REMOVED***);
***REMOVED***;

const Impl = require("../websockets/WebSocket-impl.js");
