// jsdom/lib/jsdom/living/generated/XMLHttpRequestEventTarget.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const EventHandlerNonNull = require("./EventHandlerNonNull.js");
const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;
const EventTarget = require("./EventTarget.js");

const interfaceName = "XMLHttpRequestEventTarget";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'XMLHttpRequestEventTarget'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["XMLHttpRequestEventTarget"];
  if (ctor === undefined) {
    throw new Error(
      "Internal error: constructor XMLHttpRequestEventTarget is not installed on the passed global object"
    );
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {
  EventTarget._internalSetup(wrapper, globalObject);
***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window", "DedicatedWorker", "SharedWorker"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***

  if (globalObject.EventTarget === undefined) {
    throw new Error("Internal error: attempting to evaluate XMLHttpRequestEventTarget before EventTarget");
  ***REMOVED***
  class XMLHttpRequestEventTarget extends globalObject.EventTarget {
    constructor() {
      throw new TypeError("Illegal constructor");
    ***REMOVED***

    get onloadstart() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onloadstart' called on an object that is not a valid instance of XMLHttpRequestEventTarget."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onloadstart"]);
    ***REMOVED***

    set onloadstart(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onloadstart' called on an object that is not a valid instance of XMLHttpRequestEventTarget."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onloadstart' property on 'XMLHttpRequestEventTarget': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onloadstart"] = V;
    ***REMOVED***

    get onprogress() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onprogress' called on an object that is not a valid instance of XMLHttpRequestEventTarget."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onprogress"]);
    ***REMOVED***

    set onprogress(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onprogress' called on an object that is not a valid instance of XMLHttpRequestEventTarget."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onprogress' property on 'XMLHttpRequestEventTarget': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onprogress"] = V;
    ***REMOVED***

    get onabort() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onabort' called on an object that is not a valid instance of XMLHttpRequestEventTarget."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onabort"]);
    ***REMOVED***

    set onabort(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onabort' called on an object that is not a valid instance of XMLHttpRequestEventTarget."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onabort' property on 'XMLHttpRequestEventTarget': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onabort"] = V;
    ***REMOVED***

    get onerror() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onerror' called on an object that is not a valid instance of XMLHttpRequestEventTarget."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onerror"]);
    ***REMOVED***

    set onerror(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onerror' called on an object that is not a valid instance of XMLHttpRequestEventTarget."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onerror' property on 'XMLHttpRequestEventTarget': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onerror"] = V;
    ***REMOVED***

    get onload() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onload' called on an object that is not a valid instance of XMLHttpRequestEventTarget."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onload"]);
    ***REMOVED***

    set onload(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onload' called on an object that is not a valid instance of XMLHttpRequestEventTarget."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onload' property on 'XMLHttpRequestEventTarget': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onload"] = V;
    ***REMOVED***

    get ontimeout() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get ontimeout' called on an object that is not a valid instance of XMLHttpRequestEventTarget."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["ontimeout"]);
    ***REMOVED***

    set ontimeout(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set ontimeout' called on an object that is not a valid instance of XMLHttpRequestEventTarget."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'ontimeout' property on 'XMLHttpRequestEventTarget': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["ontimeout"] = V;
    ***REMOVED***

    get onloadend() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onloadend' called on an object that is not a valid instance of XMLHttpRequestEventTarget."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onloadend"]);
    ***REMOVED***

    set onloadend(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onloadend' called on an object that is not a valid instance of XMLHttpRequestEventTarget."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onloadend' property on 'XMLHttpRequestEventTarget': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onloadend"] = V;
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(XMLHttpRequestEventTarget.prototype, {
    onloadstart: { enumerable: true ***REMOVED***,
    onprogress: { enumerable: true ***REMOVED***,
    onabort: { enumerable: true ***REMOVED***,
    onerror: { enumerable: true ***REMOVED***,
    onload: { enumerable: true ***REMOVED***,
    ontimeout: { enumerable: true ***REMOVED***,
    onloadend: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "XMLHttpRequestEventTarget", configurable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = XMLHttpRequestEventTarget;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: XMLHttpRequestEventTarget
  ***REMOVED***);
***REMOVED***;

const Impl = require("../xhr/XMLHttpRequestEventTarget-impl.js");
