// jsdom/lib/jsdom/living/generated/XMLHttpRequest.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const Document = require("./Document.js");
const Blob = require("./Blob.js");
const FormData = require("./FormData.js");
const EventHandlerNonNull = require("./EventHandlerNonNull.js");
const XMLHttpRequestResponseType = require("./XMLHttpRequestResponseType.js");
const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;
const XMLHttpRequestEventTarget = require("./XMLHttpRequestEventTarget.js");

const interfaceName = "XMLHttpRequest";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
***REMOVED***;
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
***REMOVED***;
exports.convert = (value, { context = "The provided value" ***REMOVED*** = {***REMOVED***) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  ***REMOVED***
  throw new TypeError(`${context***REMOVED*** is not of type 'XMLHttpRequest'.`);
***REMOVED***;

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  ***REMOVED***

  const ctor = globalObject[ctorRegistrySymbol]["XMLHttpRequest"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor XMLHttpRequest is not installed on the passed global object");
  ***REMOVED***

  return Object.create(ctor.prototype);
***REMOVED***

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
***REMOVED***;

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
***REMOVED***;

exports._internalSetup = (wrapper, globalObject) => {
  XMLHttpRequestEventTarget._internalSetup(wrapper, globalObject);
***REMOVED***;

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {***REMOVED***) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper;
***REMOVED***;

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  ***REMOVED***);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  ***REMOVED***
  return wrapper[implSymbol];
***REMOVED***;

const exposed = new Set(["Window", "DedicatedWorker", "SharedWorker"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  ***REMOVED***

  if (globalObject.XMLHttpRequestEventTarget === undefined) {
    throw new Error("Internal error: attempting to evaluate XMLHttpRequest before XMLHttpRequestEventTarget");
  ***REMOVED***
  class XMLHttpRequest extends globalObject.XMLHttpRequestEventTarget {
    constructor() {
      return exports.setup(Object.create(new.target.prototype), globalObject, undefined);
    ***REMOVED***

    open(method, url) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'open' called on an object that is not a valid instance of XMLHttpRequest.");
      ***REMOVED***

      if (arguments.length < 2) {
        throw new TypeError(
          "Failed to execute 'open' on 'XMLHttpRequest': 2 arguments required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      switch (arguments.length) {
        case 2:
          {
            let curArg = arguments[0];
            curArg = conversions["ByteString"](curArg, {
              context: "Failed to execute 'open' on 'XMLHttpRequest': parameter 1"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[1];
            curArg = conversions["USVString"](curArg, {
              context: "Failed to execute 'open' on 'XMLHttpRequest': parameter 2"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          break;
        case 3:
          {
            let curArg = arguments[0];
            curArg = conversions["ByteString"](curArg, {
              context: "Failed to execute 'open' on 'XMLHttpRequest': parameter 1"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[1];
            curArg = conversions["USVString"](curArg, {
              context: "Failed to execute 'open' on 'XMLHttpRequest': parameter 2"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[2];
            curArg = conversions["boolean"](curArg, {
              context: "Failed to execute 'open' on 'XMLHttpRequest': parameter 3"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          break;
        case 4:
          {
            let curArg = arguments[0];
            curArg = conversions["ByteString"](curArg, {
              context: "Failed to execute 'open' on 'XMLHttpRequest': parameter 1"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[1];
            curArg = conversions["USVString"](curArg, {
              context: "Failed to execute 'open' on 'XMLHttpRequest': parameter 2"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[2];
            curArg = conversions["boolean"](curArg, {
              context: "Failed to execute 'open' on 'XMLHttpRequest': parameter 3"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[3];
            if (curArg !== undefined) {
              if (curArg === null || curArg === undefined) {
                curArg = null;
              ***REMOVED*** else {
                curArg = conversions["USVString"](curArg, {
                  context: "Failed to execute 'open' on 'XMLHttpRequest': parameter 4"
                ***REMOVED***);
              ***REMOVED***
            ***REMOVED*** else {
              curArg = null;
            ***REMOVED***
            args.push(curArg);
          ***REMOVED***
          break;
        default:
          {
            let curArg = arguments[0];
            curArg = conversions["ByteString"](curArg, {
              context: "Failed to execute 'open' on 'XMLHttpRequest': parameter 1"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[1];
            curArg = conversions["USVString"](curArg, {
              context: "Failed to execute 'open' on 'XMLHttpRequest': parameter 2"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[2];
            curArg = conversions["boolean"](curArg, {
              context: "Failed to execute 'open' on 'XMLHttpRequest': parameter 3"
            ***REMOVED***);
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[3];
            if (curArg !== undefined) {
              if (curArg === null || curArg === undefined) {
                curArg = null;
              ***REMOVED*** else {
                curArg = conversions["USVString"](curArg, {
                  context: "Failed to execute 'open' on 'XMLHttpRequest': parameter 4"
                ***REMOVED***);
              ***REMOVED***
            ***REMOVED*** else {
              curArg = null;
            ***REMOVED***
            args.push(curArg);
          ***REMOVED***
          {
            let curArg = arguments[4];
            if (curArg !== undefined) {
              if (curArg === null || curArg === undefined) {
                curArg = null;
              ***REMOVED*** else {
                curArg = conversions["USVString"](curArg, {
                  context: "Failed to execute 'open' on 'XMLHttpRequest': parameter 5"
                ***REMOVED***);
              ***REMOVED***
            ***REMOVED*** else {
              curArg = null;
            ***REMOVED***
            args.push(curArg);
          ***REMOVED***
      ***REMOVED***
      return esValue[implSymbol].open(...args);
    ***REMOVED***

    setRequestHeader(name, value) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'setRequestHeader' called on an object that is not a valid instance of XMLHttpRequest.");
      ***REMOVED***

      if (arguments.length < 2) {
        throw new TypeError(
          "Failed to execute 'setRequestHeader' on 'XMLHttpRequest': 2 arguments required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["ByteString"](curArg, {
          context: "Failed to execute 'setRequestHeader' on 'XMLHttpRequest': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      {
        let curArg = arguments[1];
        curArg = conversions["ByteString"](curArg, {
          context: "Failed to execute 'setRequestHeader' on 'XMLHttpRequest': parameter 2"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].setRequestHeader(...args);
    ***REMOVED***

    send() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'send' called on an object that is not a valid instance of XMLHttpRequest.");
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        if (curArg !== undefined) {
          if (curArg === null || curArg === undefined) {
            curArg = null;
          ***REMOVED*** else {
            if (Document.is(curArg) || Blob.is(curArg) || FormData.is(curArg)) {
              curArg = utils.implForWrapper(curArg);
            ***REMOVED*** else if (utils.isArrayBuffer(curArg)) {
            ***REMOVED*** else if (ArrayBuffer.isView(curArg)) {
            ***REMOVED*** else {
              curArg = conversions["USVString"](curArg, {
                context: "Failed to execute 'send' on 'XMLHttpRequest': parameter 1"
              ***REMOVED***);
            ***REMOVED***
          ***REMOVED***
        ***REMOVED*** else {
          curArg = null;
        ***REMOVED***
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].send(...args);
    ***REMOVED***

    abort() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'abort' called on an object that is not a valid instance of XMLHttpRequest.");
      ***REMOVED***

      return esValue[implSymbol].abort();
    ***REMOVED***

    getResponseHeader(name) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'getResponseHeader' called on an object that is not a valid instance of XMLHttpRequest.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'getResponseHeader' on 'XMLHttpRequest': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["ByteString"](curArg, {
          context: "Failed to execute 'getResponseHeader' on 'XMLHttpRequest': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].getResponseHeader(...args);
    ***REMOVED***

    getAllResponseHeaders() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError(
          "'getAllResponseHeaders' called on an object that is not a valid instance of XMLHttpRequest."
        );
      ***REMOVED***

      return esValue[implSymbol].getAllResponseHeaders();
    ***REMOVED***

    overrideMimeType(mime) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'overrideMimeType' called on an object that is not a valid instance of XMLHttpRequest.");
      ***REMOVED***

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'overrideMimeType' on 'XMLHttpRequest': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      ***REMOVED***
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'overrideMimeType' on 'XMLHttpRequest': parameter 1"
        ***REMOVED***);
        args.push(curArg);
      ***REMOVED***
      return esValue[implSymbol].overrideMimeType(...args);
    ***REMOVED***

    get onreadystatechange() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get onreadystatechange' called on an object that is not a valid instance of XMLHttpRequest."
        );
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["onreadystatechange"]);
    ***REMOVED***

    set onreadystatechange(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set onreadystatechange' called on an object that is not a valid instance of XMLHttpRequest."
        );
      ***REMOVED***

      if (!utils.isObject(V)) {
        V = null;
      ***REMOVED*** else {
        V = EventHandlerNonNull.convert(V, {
          context: "Failed to set the 'onreadystatechange' property on 'XMLHttpRequest': The provided value"
        ***REMOVED***);
      ***REMOVED***
      esValue[implSymbol]["onreadystatechange"] = V;
    ***REMOVED***

    get readyState() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get readyState' called on an object that is not a valid instance of XMLHttpRequest.");
      ***REMOVED***

      return esValue[implSymbol]["readyState"];
    ***REMOVED***

    get timeout() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get timeout' called on an object that is not a valid instance of XMLHttpRequest.");
      ***REMOVED***

      return esValue[implSymbol]["timeout"];
    ***REMOVED***

    set timeout(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set timeout' called on an object that is not a valid instance of XMLHttpRequest.");
      ***REMOVED***

      V = conversions["unsigned long"](V, {
        context: "Failed to set the 'timeout' property on 'XMLHttpRequest': The provided value"
      ***REMOVED***);

      esValue[implSymbol]["timeout"] = V;
    ***REMOVED***

    get withCredentials() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'get withCredentials' called on an object that is not a valid instance of XMLHttpRequest."
        );
      ***REMOVED***

      return esValue[implSymbol]["withCredentials"];
    ***REMOVED***

    set withCredentials(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError(
          "'set withCredentials' called on an object that is not a valid instance of XMLHttpRequest."
        );
      ***REMOVED***

      V = conversions["boolean"](V, {
        context: "Failed to set the 'withCredentials' property on 'XMLHttpRequest': The provided value"
      ***REMOVED***);

      esValue[implSymbol]["withCredentials"] = V;
    ***REMOVED***

    get upload() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get upload' called on an object that is not a valid instance of XMLHttpRequest.");
      ***REMOVED***

      return utils.getSameObject(this, "upload", () => {
        return utils.tryWrapperForImpl(esValue[implSymbol]["upload"]);
      ***REMOVED***);
    ***REMOVED***

    get responseURL() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get responseURL' called on an object that is not a valid instance of XMLHttpRequest.");
      ***REMOVED***

      return esValue[implSymbol]["responseURL"];
    ***REMOVED***

    get status() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get status' called on an object that is not a valid instance of XMLHttpRequest.");
      ***REMOVED***

      return esValue[implSymbol]["status"];
    ***REMOVED***

    get statusText() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get statusText' called on an object that is not a valid instance of XMLHttpRequest.");
      ***REMOVED***

      return esValue[implSymbol]["statusText"];
    ***REMOVED***

    get responseType() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get responseType' called on an object that is not a valid instance of XMLHttpRequest.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["responseType"]);
    ***REMOVED***

    set responseType(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set responseType' called on an object that is not a valid instance of XMLHttpRequest.");
      ***REMOVED***

      V = `${V***REMOVED***`;
      if (!XMLHttpRequestResponseType.enumerationValues.has(V)) {
        return;
      ***REMOVED***

      esValue[implSymbol]["responseType"] = V;
    ***REMOVED***

    get response() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get response' called on an object that is not a valid instance of XMLHttpRequest.");
      ***REMOVED***

      return esValue[implSymbol]["response"];
    ***REMOVED***

    get responseText() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get responseText' called on an object that is not a valid instance of XMLHttpRequest.");
      ***REMOVED***

      return esValue[implSymbol]["responseText"];
    ***REMOVED***

    get responseXML() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get responseXML' called on an object that is not a valid instance of XMLHttpRequest.");
      ***REMOVED***

      return utils.tryWrapperForImpl(esValue[implSymbol]["responseXML"]);
    ***REMOVED***
  ***REMOVED***
  Object.defineProperties(XMLHttpRequest.prototype, {
    open: { enumerable: true ***REMOVED***,
    setRequestHeader: { enumerable: true ***REMOVED***,
    send: { enumerable: true ***REMOVED***,
    abort: { enumerable: true ***REMOVED***,
    getResponseHeader: { enumerable: true ***REMOVED***,
    getAllResponseHeaders: { enumerable: true ***REMOVED***,
    overrideMimeType: { enumerable: true ***REMOVED***,
    onreadystatechange: { enumerable: true ***REMOVED***,
    readyState: { enumerable: true ***REMOVED***,
    timeout: { enumerable: true ***REMOVED***,
    withCredentials: { enumerable: true ***REMOVED***,
    upload: { enumerable: true ***REMOVED***,
    responseURL: { enumerable: true ***REMOVED***,
    status: { enumerable: true ***REMOVED***,
    statusText: { enumerable: true ***REMOVED***,
    responseType: { enumerable: true ***REMOVED***,
    response: { enumerable: true ***REMOVED***,
    responseText: { enumerable: true ***REMOVED***,
    responseXML: { enumerable: true ***REMOVED***,
    [Symbol.toStringTag]: { value: "XMLHttpRequest", configurable: true ***REMOVED***,
    UNSENT: { value: 0, enumerable: true ***REMOVED***,
    OPENED: { value: 1, enumerable: true ***REMOVED***,
    HEADERS_RECEIVED: { value: 2, enumerable: true ***REMOVED***,
    LOADING: { value: 3, enumerable: true ***REMOVED***,
    DONE: { value: 4, enumerable: true ***REMOVED***
  ***REMOVED***);
  Object.defineProperties(XMLHttpRequest, {
    UNSENT: { value: 0, enumerable: true ***REMOVED***,
    OPENED: { value: 1, enumerable: true ***REMOVED***,
    HEADERS_RECEIVED: { value: 2, enumerable: true ***REMOVED***,
    LOADING: { value: 3, enumerable: true ***REMOVED***,
    DONE: { value: 4, enumerable: true ***REMOVED***
  ***REMOVED***);
  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  ***REMOVED***
  globalObject[ctorRegistrySymbol][interfaceName] = XMLHttpRequest;

  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: XMLHttpRequest
  ***REMOVED***);
***REMOVED***;

const Impl = require("../xhr/XMLHttpRequest-impl.js");
