// ajv/dist/vocabularies/validation/limitLength.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const codegen_1 = require("../../compile/codegen");
const util_1 = require("../../compile/util");
const ucs2length_1 = require("../../runtime/ucs2length");
const error = {
    message({ keyword, schemaCode ***REMOVED***) {
        const comp = keyword === "maxLength" ? "more" : "fewer";
        return (0, codegen_1.str) `must NOT have ${comp***REMOVED*** than ${schemaCode***REMOVED*** characters`;
    ***REMOVED***,
    params: ({ schemaCode ***REMOVED***) => (0, codegen_1._) `{limit: ${schemaCode***REMOVED******REMOVED***`,
***REMOVED***;
const def = {
    keyword: ["maxLength", "minLength"],
    type: "string",
    schemaType: "number",
    $data: true,
    error,
    code(cxt) {
        const { keyword, data, schemaCode, it ***REMOVED*** = cxt;
        const op = keyword === "maxLength" ? codegen_1.operators.GT : codegen_1.operators.LT;
        const len = it.opts.unicode === false ? (0, codegen_1._) `${data***REMOVED***.length` : (0, codegen_1._) `${(0, util_1.useFunc)(cxt.gen, ucs2length_1.default)***REMOVED***(${data***REMOVED***)`;
        cxt.fail$data((0, codegen_1._) `${len***REMOVED*** ${op***REMOVED*** ${schemaCode***REMOVED***`);
    ***REMOVED***,
***REMOVED***;
exports.default = def;
