// whatwg-url/lib/URL.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const impl = utils.implSymbol;

class URL {
  constructor(url) {
    if (arguments.length < 1) {
      throw new TypeError("Failed to construct 'URL': 1 argument required, but only " + arguments.length + " present.");
    ***REMOVED***
    const args = [];
    {
      let curArg = arguments[0];
      curArg = conversions["USVString"](curArg, { context: "Failed to construct 'URL': parameter 1" ***REMOVED***);
      args.push(curArg);
    ***REMOVED***
    {
      let curArg = arguments[1];
      if (curArg !== undefined) {
        curArg = conversions["USVString"](curArg, { context: "Failed to construct 'URL': parameter 2" ***REMOVED***);
      ***REMOVED***
      args.push(curArg);
    ***REMOVED***
    return iface.setup(Object.create(new.target.prototype), args);
  ***REMOVED***

  toJSON() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    return this[impl].toJSON();
  ***REMOVED***

  get href() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    return this[impl]["href"];
  ***REMOVED***

  set href(V) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    V = conversions["USVString"](V, { context: "Failed to set the 'href' property on 'URL': The provided value" ***REMOVED***);

    this[impl]["href"] = V;
  ***REMOVED***

  toString() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***
    return this[impl]["href"];
  ***REMOVED***

  get origin() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    return this[impl]["origin"];
  ***REMOVED***

  get protocol() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    return this[impl]["protocol"];
  ***REMOVED***

  set protocol(V) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    V = conversions["USVString"](V, { context: "Failed to set the 'protocol' property on 'URL': The provided value" ***REMOVED***);

    this[impl]["protocol"] = V;
  ***REMOVED***

  get username() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    return this[impl]["username"];
  ***REMOVED***

  set username(V) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    V = conversions["USVString"](V, { context: "Failed to set the 'username' property on 'URL': The provided value" ***REMOVED***);

    this[impl]["username"] = V;
  ***REMOVED***

  get password() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    return this[impl]["password"];
  ***REMOVED***

  set password(V) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    V = conversions["USVString"](V, { context: "Failed to set the 'password' property on 'URL': The provided value" ***REMOVED***);

    this[impl]["password"] = V;
  ***REMOVED***

  get host() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    return this[impl]["host"];
  ***REMOVED***

  set host(V) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    V = conversions["USVString"](V, { context: "Failed to set the 'host' property on 'URL': The provided value" ***REMOVED***);

    this[impl]["host"] = V;
  ***REMOVED***

  get hostname() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    return this[impl]["hostname"];
  ***REMOVED***

  set hostname(V) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    V = conversions["USVString"](V, { context: "Failed to set the 'hostname' property on 'URL': The provided value" ***REMOVED***);

    this[impl]["hostname"] = V;
  ***REMOVED***

  get port() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    return this[impl]["port"];
  ***REMOVED***

  set port(V) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    V = conversions["USVString"](V, { context: "Failed to set the 'port' property on 'URL': The provided value" ***REMOVED***);

    this[impl]["port"] = V;
  ***REMOVED***

  get pathname() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    return this[impl]["pathname"];
  ***REMOVED***

  set pathname(V) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    V = conversions["USVString"](V, { context: "Failed to set the 'pathname' property on 'URL': The provided value" ***REMOVED***);

    this[impl]["pathname"] = V;
  ***REMOVED***

  get search() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    return this[impl]["search"];
  ***REMOVED***

  set search(V) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    V = conversions["USVString"](V, { context: "Failed to set the 'search' property on 'URL': The provided value" ***REMOVED***);

    this[impl]["search"] = V;
  ***REMOVED***

  get searchParams() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    return utils.getSameObject(this, "searchParams", () => {
      return utils.tryWrapperForImpl(this[impl]["searchParams"]);
    ***REMOVED***);
  ***REMOVED***

  get hash() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    return this[impl]["hash"];
  ***REMOVED***

  set hash(V) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    V = conversions["USVString"](V, { context: "Failed to set the 'hash' property on 'URL': The provided value" ***REMOVED***);

    this[impl]["hash"] = V;
  ***REMOVED***
***REMOVED***
Object.defineProperties(URL.prototype, {
  toJSON: { enumerable: true ***REMOVED***,
  href: { enumerable: true ***REMOVED***,
  toString: { enumerable: true ***REMOVED***,
  origin: { enumerable: true ***REMOVED***,
  protocol: { enumerable: true ***REMOVED***,
  username: { enumerable: true ***REMOVED***,
  password: { enumerable: true ***REMOVED***,
  host: { enumerable: true ***REMOVED***,
  hostname: { enumerable: true ***REMOVED***,
  port: { enumerable: true ***REMOVED***,
  pathname: { enumerable: true ***REMOVED***,
  search: { enumerable: true ***REMOVED***,
  searchParams: { enumerable: true ***REMOVED***,
  hash: { enumerable: true ***REMOVED***,
  [Symbol.toStringTag]: { value: "URL", configurable: true ***REMOVED***
***REMOVED***);
const iface = {
  // When an interface-module that implements this interface as a mixin is loaded, it will append its own `.is()`
  // method into this array. It allows objects that directly implements *those* interfaces to be recognized as
  // implementing this mixin interface.
  _mixedIntoPredicates: [],
  is(obj) {
    if (obj) {
      if (utils.hasOwn(obj, impl) && obj[impl] instanceof Impl.implementation) {
        return true;
      ***REMOVED***
      for (const isMixedInto of module.exports._mixedIntoPredicates) {
        if (isMixedInto(obj)) {
          return true;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
    return false;
  ***REMOVED***,
  isImpl(obj) {
    if (obj) {
      if (obj instanceof Impl.implementation) {
        return true;
      ***REMOVED***

      const wrapper = utils.wrapperForImpl(obj);
      for (const isMixedInto of module.exports._mixedIntoPredicates) {
        if (isMixedInto(wrapper)) {
          return true;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
    return false;
  ***REMOVED***,
  convert(obj, { context = "The provided value" ***REMOVED*** = {***REMOVED***) {
    if (module.exports.is(obj)) {
      return utils.implForWrapper(obj);
    ***REMOVED***
    throw new TypeError(`${context***REMOVED*** is not of type 'URL'.`);
  ***REMOVED***,

  create(constructorArgs, privateData) {
    let obj = Object.create(URL.prototype);
    obj = this.setup(obj, constructorArgs, privateData);
    return obj;
  ***REMOVED***,
  createImpl(constructorArgs, privateData) {
    let obj = Object.create(URL.prototype);
    obj = this.setup(obj, constructorArgs, privateData);
    return utils.implForWrapper(obj);
  ***REMOVED***,
  _internalSetup(obj) {***REMOVED***,
  setup(obj, constructorArgs, privateData) {
    if (!privateData) privateData = {***REMOVED***;

    privateData.wrapper = obj;

    this._internalSetup(obj);
    Object.defineProperty(obj, impl, {
      value: new Impl.implementation(constructorArgs, privateData),
      configurable: true
    ***REMOVED***);

    obj[impl][utils.wrapperSymbol] = obj;
    if (Impl.init) {
      Impl.init(obj[impl], privateData);
    ***REMOVED***
    return obj;
  ***REMOVED***,
  interface: URL,
  expose: {
    Window: { URL ***REMOVED***,
    Worker: { URL ***REMOVED***
  ***REMOVED***
***REMOVED***; // iface
module.exports = iface;

const Impl = require("./URL-impl.js");
