// whatwg-url/lib/URLSearchParams.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const impl = utils.implSymbol;

const IteratorPrototype = Object.create(utils.IteratorPrototype, {
  next: {
    value: function next() {
      const internal = this[utils.iterInternalSymbol];
      const { target, kind, index ***REMOVED*** = internal;
      const values = Array.from(target[impl]);
      const len = values.length;
      if (index >= len) {
        return { value: undefined, done: true ***REMOVED***;
      ***REMOVED***

      const pair = values[index];
      internal.index = index + 1;
      const [key, value] = pair.map(utils.tryWrapperForImpl);

      let result;
      switch (kind) {
        case "key":
          result = key;
          break;
        case "value":
          result = value;
          break;
        case "key+value":
          result = [key, value];
          break;
      ***REMOVED***
      return { value: result, done: false ***REMOVED***;
    ***REMOVED***,
    writable: true,
    enumerable: true,
    configurable: true
  ***REMOVED***,
  [Symbol.toStringTag]: {
    value: "URLSearchParams Iterator",
    configurable: true
  ***REMOVED***
***REMOVED***);
class URLSearchParams {
  constructor() {
    const args = [];
    {
      let curArg = arguments[0];
      if (curArg !== undefined) {
        if (utils.isObject(curArg)) {
          if (curArg[Symbol.iterator] !== undefined) {
            if (!utils.isObject(curArg)) {
              throw new TypeError(
                "Failed to construct 'URLSearchParams': parameter 1" + " sequence" + " is not an iterable object."
              );
            ***REMOVED*** else {
              const V = [];
              const tmp = curArg;
              for (let nextItem of tmp) {
                if (!utils.isObject(nextItem)) {
                  throw new TypeError(
                    "Failed to construct 'URLSearchParams': parameter 1" +
                      " sequence" +
                      "'s element" +
                      " is not an iterable object."
                  );
                ***REMOVED*** else {
                  const V = [];
                  const tmp = nextItem;
                  for (let nextItem of tmp) {
                    nextItem = conversions["USVString"](nextItem, {
                      context:
                        "Failed to construct 'URLSearchParams': parameter 1" + " sequence" + "'s element" + "'s element"
                    ***REMOVED***);

                    V.push(nextItem);
                  ***REMOVED***
                  nextItem = V;
                ***REMOVED***

                V.push(nextItem);
              ***REMOVED***
              curArg = V;
            ***REMOVED***
          ***REMOVED*** else {
            if (!utils.isObject(curArg)) {
              throw new TypeError(
                "Failed to construct 'URLSearchParams': parameter 1" + " record" + " is not an object."
              );
            ***REMOVED*** else {
              const result = Object.create(null);
              for (const key of Reflect.ownKeys(curArg)) {
                const desc = Object.getOwnPropertyDescriptor(curArg, key);
                if (desc && desc.enumerable) {
                  let typedKey = key;
                  let typedValue = curArg[key];

                  typedKey = conversions["USVString"](typedKey, {
                    context: "Failed to construct 'URLSearchParams': parameter 1" + " record" + "'s key"
                  ***REMOVED***);

                  typedValue = conversions["USVString"](typedValue, {
                    context: "Failed to construct 'URLSearchParams': parameter 1" + " record" + "'s value"
                  ***REMOVED***);

                  result[typedKey] = typedValue;
                ***REMOVED***
              ***REMOVED***
              curArg = result;
            ***REMOVED***
          ***REMOVED***
        ***REMOVED*** else {
          curArg = conversions["USVString"](curArg, { context: "Failed to construct 'URLSearchParams': parameter 1" ***REMOVED***);
        ***REMOVED***
      ***REMOVED*** else {
        curArg = "";
      ***REMOVED***
      args.push(curArg);
    ***REMOVED***
    return iface.setup(Object.create(new.target.prototype), args);
  ***REMOVED***

  append(name, value) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    if (arguments.length < 2) {
      throw new TypeError(
        "Failed to execute 'append' on 'URLSearchParams': 2 arguments required, but only " +
          arguments.length +
          " present."
      );
    ***REMOVED***
    const args = [];
    {
      let curArg = arguments[0];
      curArg = conversions["USVString"](curArg, {
        context: "Failed to execute 'append' on 'URLSearchParams': parameter 1"
      ***REMOVED***);
      args.push(curArg);
    ***REMOVED***
    {
      let curArg = arguments[1];
      curArg = conversions["USVString"](curArg, {
        context: "Failed to execute 'append' on 'URLSearchParams': parameter 2"
      ***REMOVED***);
      args.push(curArg);
    ***REMOVED***
    return this[impl].append(...args);
  ***REMOVED***

  delete(name) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    if (arguments.length < 1) {
      throw new TypeError(
        "Failed to execute 'delete' on 'URLSearchParams': 1 argument required, but only " +
          arguments.length +
          " present."
      );
    ***REMOVED***
    const args = [];
    {
      let curArg = arguments[0];
      curArg = conversions["USVString"](curArg, {
        context: "Failed to execute 'delete' on 'URLSearchParams': parameter 1"
      ***REMOVED***);
      args.push(curArg);
    ***REMOVED***
    return this[impl].delete(...args);
  ***REMOVED***

  get(name) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    if (arguments.length < 1) {
      throw new TypeError(
        "Failed to execute 'get' on 'URLSearchParams': 1 argument required, but only " + arguments.length + " present."
      );
    ***REMOVED***
    const args = [];
    {
      let curArg = arguments[0];
      curArg = conversions["USVString"](curArg, {
        context: "Failed to execute 'get' on 'URLSearchParams': parameter 1"
      ***REMOVED***);
      args.push(curArg);
    ***REMOVED***
    return this[impl].get(...args);
  ***REMOVED***

  getAll(name) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    if (arguments.length < 1) {
      throw new TypeError(
        "Failed to execute 'getAll' on 'URLSearchParams': 1 argument required, but only " +
          arguments.length +
          " present."
      );
    ***REMOVED***
    const args = [];
    {
      let curArg = arguments[0];
      curArg = conversions["USVString"](curArg, {
        context: "Failed to execute 'getAll' on 'URLSearchParams': parameter 1"
      ***REMOVED***);
      args.push(curArg);
    ***REMOVED***
    return utils.tryWrapperForImpl(this[impl].getAll(...args));
  ***REMOVED***

  has(name) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    if (arguments.length < 1) {
      throw new TypeError(
        "Failed to execute 'has' on 'URLSearchParams': 1 argument required, but only " + arguments.length + " present."
      );
    ***REMOVED***
    const args = [];
    {
      let curArg = arguments[0];
      curArg = conversions["USVString"](curArg, {
        context: "Failed to execute 'has' on 'URLSearchParams': parameter 1"
      ***REMOVED***);
      args.push(curArg);
    ***REMOVED***
    return this[impl].has(...args);
  ***REMOVED***

  set(name, value) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    if (arguments.length < 2) {
      throw new TypeError(
        "Failed to execute 'set' on 'URLSearchParams': 2 arguments required, but only " + arguments.length + " present."
      );
    ***REMOVED***
    const args = [];
    {
      let curArg = arguments[0];
      curArg = conversions["USVString"](curArg, {
        context: "Failed to execute 'set' on 'URLSearchParams': parameter 1"
      ***REMOVED***);
      args.push(curArg);
    ***REMOVED***
    {
      let curArg = arguments[1];
      curArg = conversions["USVString"](curArg, {
        context: "Failed to execute 'set' on 'URLSearchParams': parameter 2"
      ***REMOVED***);
      args.push(curArg);
    ***REMOVED***
    return this[impl].set(...args);
  ***REMOVED***

  sort() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    return this[impl].sort();
  ***REMOVED***

  toString() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***

    return this[impl].toString();
  ***REMOVED***

  keys() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***
    return module.exports.createDefaultIterator(this, "key");
  ***REMOVED***

  values() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***
    return module.exports.createDefaultIterator(this, "value");
  ***REMOVED***

  entries() {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***
    return module.exports.createDefaultIterator(this, "key+value");
  ***REMOVED***

  forEach(callback) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    ***REMOVED***
    if (arguments.length < 1) {
      throw new TypeError("Failed to execute 'forEach' on 'iterable': 1 argument required, " + "but only 0 present.");
    ***REMOVED***
    if (typeof callback !== "function") {
      throw new TypeError(
        "Failed to execute 'forEach' on 'iterable': The callback provided " + "as parameter 1 is not a function."
      );
    ***REMOVED***
    const thisArg = arguments[1];
    let pairs = Array.from(this[impl]);
    let i = 0;
    while (i < pairs.length) {
      const [key, value] = pairs[i].map(utils.tryWrapperForImpl);
      callback.call(thisArg, value, key, this);
      pairs = Array.from(this[impl]);
      i++;
    ***REMOVED***
  ***REMOVED***
***REMOVED***
Object.defineProperties(URLSearchParams.prototype, {
  append: { enumerable: true ***REMOVED***,
  delete: { enumerable: true ***REMOVED***,
  get: { enumerable: true ***REMOVED***,
  getAll: { enumerable: true ***REMOVED***,
  has: { enumerable: true ***REMOVED***,
  set: { enumerable: true ***REMOVED***,
  sort: { enumerable: true ***REMOVED***,
  toString: { enumerable: true ***REMOVED***,
  keys: { enumerable: true ***REMOVED***,
  values: { enumerable: true ***REMOVED***,
  entries: { enumerable: true ***REMOVED***,
  forEach: { enumerable: true ***REMOVED***,
  [Symbol.toStringTag]: { value: "URLSearchParams", configurable: true ***REMOVED***,
  [Symbol.iterator]: { value: URLSearchParams.prototype.entries, configurable: true, writable: true ***REMOVED***
***REMOVED***);
const iface = {
  // When an interface-module that implements this interface as a mixin is loaded, it will append its own `.is()`
  // method into this array. It allows objects that directly implements *those* interfaces to be recognized as
  // implementing this mixin interface.
  _mixedIntoPredicates: [],
  is(obj) {
    if (obj) {
      if (utils.hasOwn(obj, impl) && obj[impl] instanceof Impl.implementation) {
        return true;
      ***REMOVED***
      for (const isMixedInto of module.exports._mixedIntoPredicates) {
        if (isMixedInto(obj)) {
          return true;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
    return false;
  ***REMOVED***,
  isImpl(obj) {
    if (obj) {
      if (obj instanceof Impl.implementation) {
        return true;
      ***REMOVED***

      const wrapper = utils.wrapperForImpl(obj);
      for (const isMixedInto of module.exports._mixedIntoPredicates) {
        if (isMixedInto(wrapper)) {
          return true;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
    return false;
  ***REMOVED***,
  convert(obj, { context = "The provided value" ***REMOVED*** = {***REMOVED***) {
    if (module.exports.is(obj)) {
      return utils.implForWrapper(obj);
    ***REMOVED***
    throw new TypeError(`${context***REMOVED*** is not of type 'URLSearchParams'.`);
  ***REMOVED***,

  createDefaultIterator(target, kind) {
    const iterator = Object.create(IteratorPrototype);
    Object.defineProperty(iterator, utils.iterInternalSymbol, {
      value: { target, kind, index: 0 ***REMOVED***,
      configurable: true
    ***REMOVED***);
    return iterator;
  ***REMOVED***,

  create(constructorArgs, privateData) {
    let obj = Object.create(URLSearchParams.prototype);
    obj = this.setup(obj, constructorArgs, privateData);
    return obj;
  ***REMOVED***,
  createImpl(constructorArgs, privateData) {
    let obj = Object.create(URLSearchParams.prototype);
    obj = this.setup(obj, constructorArgs, privateData);
    return utils.implForWrapper(obj);
  ***REMOVED***,
  _internalSetup(obj) {***REMOVED***,
  setup(obj, constructorArgs, privateData) {
    if (!privateData) privateData = {***REMOVED***;

    privateData.wrapper = obj;

    this._internalSetup(obj);
    Object.defineProperty(obj, impl, {
      value: new Impl.implementation(constructorArgs, privateData),
      configurable: true
    ***REMOVED***);

    obj[impl][utils.wrapperSymbol] = obj;
    if (Impl.init) {
      Impl.init(obj[impl], privateData);
    ***REMOVED***
    return obj;
  ***REMOVED***,
  interface: URLSearchParams,
  expose: {
    Window: { URLSearchParams ***REMOVED***,
    Worker: { URLSearchParams ***REMOVED***
  ***REMOVED***
***REMOVED***; // iface
module.exports = iface;

const Impl = require("./URLSearchParams-impl.js");
