// webpack/lib/AsyncDependenciesBlock.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const DependenciesBlock = require("./DependenciesBlock");
const makeSerializable = require("./util/makeSerializable");

/** @typedef {import("./ChunkGraph")***REMOVED*** ChunkGraph */
/** @typedef {import("./ChunkGroup")***REMOVED*** ChunkGroup */
/** @typedef {import("./ChunkGroup").ChunkGroupOptions***REMOVED*** ChunkGroupOptions */
/** @typedef {import("./Dependency").DependencyLocation***REMOVED*** DependencyLocation */
/** @typedef {import("./Dependency").UpdateHashContext***REMOVED*** UpdateHashContext */
/** @typedef {import("./Entrypoint").EntryOptions***REMOVED*** EntryOptions */
/** @typedef {import("./Module")***REMOVED*** Module */
/** @typedef {import("./serialization/ObjectMiddleware").ObjectDeserializerContext***REMOVED*** ObjectDeserializerContext */
/** @typedef {import("./serialization/ObjectMiddleware").ObjectSerializerContext***REMOVED*** ObjectSerializerContext */
/** @typedef {import("./util/Hash")***REMOVED*** Hash */

/** @typedef {(ChunkGroupOptions & { entryOptions?: EntryOptions ***REMOVED***) | string***REMOVED*** GroupOptions */

class AsyncDependenciesBlock extends DependenciesBlock {
	/**
	 * @param {GroupOptions | null***REMOVED*** groupOptions options for the group
	 * @param {(DependencyLocation | null)=***REMOVED*** loc the line of code
	 * @param {(string | null)=***REMOVED*** request the request
	 */
	constructor(groupOptions, loc, request) {
		super();
		if (typeof groupOptions === "string") {
			groupOptions = { name: groupOptions ***REMOVED***;
		***REMOVED*** else if (!groupOptions) {
			groupOptions = { name: undefined ***REMOVED***;
		***REMOVED***
		this.groupOptions = groupOptions;
		this.loc = loc;
		this.request = request;
		this._stringifiedGroupOptions = undefined;
	***REMOVED***

	/**
	 * @returns {string | null | undefined***REMOVED*** The name of the chunk
	 */
	get chunkName() {
		return this.groupOptions.name;
	***REMOVED***

	/**
	 * @param {string | undefined***REMOVED*** value The new chunk name
	 * @returns {void***REMOVED***
	 */
	set chunkName(value) {
		if (this.groupOptions.name !== value) {
			this.groupOptions.name = value;
			this._stringifiedGroupOptions = undefined;
		***REMOVED***
	***REMOVED***

	/**
	 * @param {Hash***REMOVED*** hash the hash used to track dependencies
	 * @param {UpdateHashContext***REMOVED*** context context
	 * @returns {void***REMOVED***
	 */
	updateHash(hash, context) {
		const { chunkGraph ***REMOVED*** = context;
		if (this._stringifiedGroupOptions === undefined) {
			this._stringifiedGroupOptions = JSON.stringify(this.groupOptions);
		***REMOVED***
		const chunkGroup = chunkGraph.getBlockChunkGroup(this);
		hash.update(
			`${this._stringifiedGroupOptions***REMOVED***${chunkGroup ? chunkGroup.id : ""***REMOVED***`
		);
		super.updateHash(hash, context);
	***REMOVED***

	/**
	 * @param {ObjectSerializerContext***REMOVED*** context context
	 */
	serialize(context) {
		const { write ***REMOVED*** = context;
		write(this.groupOptions);
		write(this.loc);
		write(this.request);
		super.serialize(context);
	***REMOVED***

	/**
	 * @param {ObjectDeserializerContext***REMOVED*** context context
	 */
	deserialize(context) {
		const { read ***REMOVED*** = context;
		this.groupOptions = read();
		this.loc = read();
		this.request = read();
		super.deserialize(context);
	***REMOVED***
***REMOVED***

makeSerializable(AsyncDependenciesBlock, "webpack/lib/AsyncDependenciesBlock");

Object.defineProperty(AsyncDependenciesBlock.prototype, "module", {
	get() {
		throw new Error(
			"module property was removed from AsyncDependenciesBlock (it's not needed)"
		);
	***REMOVED***,
	set() {
		throw new Error(
			"module property was removed from AsyncDependenciesBlock (it's not needed)"
		);
	***REMOVED***
***REMOVED***);

module.exports = AsyncDependenciesBlock;
