// webpack/lib/ConditionalInitFragment.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const { ConcatSource, PrefixSource ***REMOVED*** = require("webpack-sources");
const InitFragment = require("./InitFragment");
const Template = require("./Template");
const { mergeRuntime ***REMOVED*** = require("./util/runtime");

/** @typedef {import("webpack-sources").Source***REMOVED*** Source */
/** @typedef {import("./Generator").GenerateContext***REMOVED*** GenerateContext */
/** @typedef {import("./util/runtime").RuntimeSpec***REMOVED*** RuntimeSpec */

/**
 * @param {string***REMOVED*** condition condition
 * @param {string | Source***REMOVED*** source source
 * @returns {string | Source***REMOVED*** wrapped source
 */
const wrapInCondition = (condition, source) => {
	if (typeof source === "string") {
		return Template.asString([
			`if (${condition***REMOVED***) {`,
			Template.indent(source),
			"***REMOVED***",
			""
		]);
	***REMOVED***
	return new ConcatSource(
		`if (${condition***REMOVED***) {\n`,
		new PrefixSource("\t", source),
		"***REMOVED***\n"
	);
***REMOVED***;

/**
 * @extends {InitFragment<GenerateContext>***REMOVED***
 */
class ConditionalInitFragment extends InitFragment {
	/**
	 * @param {string | Source | undefined***REMOVED*** content the source code that will be included as initialization code
	 * @param {number***REMOVED*** stage category of initialization code (contribute to order)
	 * @param {number***REMOVED*** position position in the category (contribute to order)
	 * @param {string | undefined***REMOVED*** key unique key to avoid emitting the same initialization code twice
	 * @param {RuntimeSpec | boolean***REMOVED*** runtimeCondition in which runtime this fragment should be executed
	 * @param {string | Source=***REMOVED*** endContent the source code that will be included at the end of the module
	 */
	constructor(
		content,
		stage,
		position,
		key,
		runtimeCondition = true,
		endContent = undefined
	) {
		super(content, stage, position, key, endContent);
		this.runtimeCondition = runtimeCondition;
	***REMOVED***

	/**
	 * @param {GenerateContext***REMOVED*** context context
	 * @returns {string | Source | undefined***REMOVED*** the source code that will be included as initialization code
	 */
	getContent(context) {
		if (this.runtimeCondition === false || !this.content) return "";
		if (this.runtimeCondition === true) return this.content;
		const expr = context.runtimeTemplate.runtimeConditionExpression({
			chunkGraph: context.chunkGraph,
			runtimeRequirements: context.runtimeRequirements,
			runtime: context.runtime,
			runtimeCondition: this.runtimeCondition
		***REMOVED***);
		if (expr === "true") return this.content;
		return wrapInCondition(expr, this.content);
	***REMOVED***

	/**
	 * @param {GenerateContext***REMOVED*** context context
	 * @returns {string | Source=***REMOVED*** the source code that will be included at the end of the module
	 */
	getEndContent(context) {
		if (this.runtimeCondition === false || !this.endContent) return "";
		if (this.runtimeCondition === true) return this.endContent;
		const expr = context.runtimeTemplate.runtimeConditionExpression({
			chunkGraph: context.chunkGraph,
			runtimeRequirements: context.runtimeRequirements,
			runtime: context.runtime,
			runtimeCondition: this.runtimeCondition
		***REMOVED***);
		if (expr === "true") return this.endContent;
		return wrapInCondition(expr, this.endContent);
	***REMOVED***

	/**
	 * @param {ConditionalInitFragment***REMOVED*** other fragment to merge with
	 * @returns {ConditionalInitFragment***REMOVED*** merged fragment
	 */
	merge(other) {
		if (this.runtimeCondition === true) return this;
		if (other.runtimeCondition === true) return other;
		if (this.runtimeCondition === false) return other;
		if (other.runtimeCondition === false) return this;
		const runtimeCondition = mergeRuntime(
			this.runtimeCondition,
			other.runtimeCondition
		);
		return new ConditionalInitFragment(
			this.content,
			this.stage,
			this.position,
			this.key,
			runtimeCondition,
			this.endContent
		);
	***REMOVED***
***REMOVED***

module.exports = ConditionalInitFragment;
