// webpack/lib/CssModule.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Alexander Akait @alexander-akait
*/

"use strict";

const NormalModule = require("./NormalModule");
const makeSerializable = require("./util/makeSerializable");

/** @typedef {import("./Module")***REMOVED*** Module */
/** @typedef {import("./NormalModule").NormalModuleCreateData***REMOVED*** NormalModuleCreateData */
/** @typedef {import("./RequestShortener")***REMOVED*** RequestShortener */
/** @typedef {import("./serialization/ObjectMiddleware").ObjectDeserializerContext***REMOVED*** ObjectDeserializerContext */
/** @typedef {import("./serialization/ObjectMiddleware").ObjectSerializerContext***REMOVED*** ObjectSerializerContext */

/** @typedef {string | undefined***REMOVED*** CssLayer */
/** @typedef {string | undefined***REMOVED*** Supports */
/** @typedef {string | undefined***REMOVED*** Media */
/** @typedef {[CssLayer, Supports, Media]***REMOVED*** InheritanceItem */
/** @typedef {Array<InheritanceItem>***REMOVED*** Inheritance */

/** @typedef {NormalModuleCreateData & { cssLayer: CssLayer, supports: Supports, media: Media, inheritance: Inheritance ***REMOVED******REMOVED*** CSSModuleCreateData */

class CssModule extends NormalModule {
	/**
	 * @param {CSSModuleCreateData***REMOVED*** options options object
	 */
	constructor(options) {
		super(options);

		// Avoid override `layer` for `Module` class, because it is a feature to run module in specific layer
		this.cssLayer = options.cssLayer;
		this.supports = options.supports;
		this.media = options.media;
		this.inheritance = options.inheritance;
	***REMOVED***

	/**
	 * @returns {string***REMOVED*** a unique identifier of the module
	 */
	identifier() {
		let identifier = super.identifier();

		if (this.cssLayer) {
			identifier += `|${this.cssLayer***REMOVED***`;
		***REMOVED***

		if (this.supports) {
			identifier += `|${this.supports***REMOVED***`;
		***REMOVED***

		if (this.media) {
			identifier += `|${this.media***REMOVED***`;
		***REMOVED***

		if (this.inheritance) {
			const inheritance = this.inheritance.map(
				(item, index) =>
					`inheritance_${index***REMOVED***|${item[0] || ""***REMOVED***|${item[1] || ""***REMOVED***|${
						item[2] || ""
					***REMOVED***`
			);

			identifier += `|${inheritance.join("|")***REMOVED***`;
		***REMOVED***

		// We generate extra code for HMR, so we need to invalidate the module
		if (this.hot) {
			identifier += `|${this.hot***REMOVED***`;
		***REMOVED***

		return identifier;
	***REMOVED***

	/**
	 * @param {RequestShortener***REMOVED*** requestShortener the request shortener
	 * @returns {string***REMOVED*** a user readable identifier of the module
	 */
	readableIdentifier(requestShortener) {
		const readableIdentifier = super.readableIdentifier(requestShortener);

		let identifier = `css ${readableIdentifier***REMOVED***`;

		if (this.cssLayer) {
			identifier += ` (layer: ${this.cssLayer***REMOVED***)`;
		***REMOVED***

		if (this.supports) {
			identifier += ` (supports: ${this.supports***REMOVED***)`;
		***REMOVED***

		if (this.media) {
			identifier += ` (media: ${this.media***REMOVED***)`;
		***REMOVED***

		return identifier;
	***REMOVED***

	/**
	 * Assuming this module is in the cache. Update the (cached) module with
	 * the fresh module from the factory. Usually updates internal references
	 * and properties.
	 * @param {Module***REMOVED*** module fresh module
	 * @returns {void***REMOVED***
	 */
	updateCacheModule(module) {
		super.updateCacheModule(module);
		const m = /** @type {CssModule***REMOVED*** */ (module);
		this.cssLayer = m.cssLayer;
		this.supports = m.supports;
		this.media = m.media;
		this.inheritance = m.inheritance;
	***REMOVED***

	/**
	 * @param {ObjectSerializerContext***REMOVED*** context context
	 */
	serialize(context) {
		const { write ***REMOVED*** = context;
		write(this.cssLayer);
		write(this.supports);
		write(this.media);
		write(this.inheritance);
		super.serialize(context);
	***REMOVED***

	/**
	 * @param {ObjectDeserializerContext***REMOVED*** context context
	 * @returns {CssModule***REMOVED*** the deserialized object
	 */
	static deserialize(context) {
		const obj = new CssModule({
			// will be deserialized by Module
			layer: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			type: "",
			// will be filled by updateCacheModule
			resource: "",
			context: "",
			request: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			userRequest: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			rawRequest: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			loaders: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			matchResource: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			parser: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			parserOptions: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			generator: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			generatorOptions: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			resolveOptions: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			cssLayer: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			supports: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			media: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			inheritance: /** @type {EXPECTED_ANY***REMOVED*** */ (null)
		***REMOVED***);
		obj.deserialize(context);
		return obj;
	***REMOVED***

	/**
	 * @param {ObjectDeserializerContext***REMOVED*** context context
	 * @returns {TODO***REMOVED*** Module
	 */
	deserialize(context) {
		const { read ***REMOVED*** = context;
		this.cssLayer = read();
		this.supports = read();
		this.media = read();
		this.inheritance = read();
		super.deserialize(context);
	***REMOVED***
***REMOVED***

makeSerializable(CssModule, "webpack/lib/CssModule");

module.exports = CssModule;
