// webpack/lib/DefinePlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const {
	JAVASCRIPT_MODULE_TYPE_AUTO,
	JAVASCRIPT_MODULE_TYPE_ESM,
	JAVASCRIPT_MODULE_TYPE_DYNAMIC
***REMOVED*** = require("./ModuleTypeConstants");
const RuntimeGlobals = require("./RuntimeGlobals");
const WebpackError = require("./WebpackError");
const ConstDependency = require("./dependencies/ConstDependency");
const BasicEvaluatedExpression = require("./javascript/BasicEvaluatedExpression");
const { VariableInfo ***REMOVED*** = require("./javascript/JavascriptParser");
const {
	evaluateToString,
	toConstantDependency
***REMOVED*** = require("./javascript/JavascriptParserHelpers");
const createHash = require("./util/createHash");

/** @typedef {import("estree").Expression***REMOVED*** Expression */
/** @typedef {import("../declarations/WebpackOptions").HashFunction***REMOVED*** HashFunction */
/** @typedef {import("./Compiler")***REMOVED*** Compiler */
/** @typedef {import("./Module").BuildInfo***REMOVED*** BuildInfo */
/** @typedef {import("./Module").ValueCacheVersions***REMOVED*** ValueCacheVersions */
/** @typedef {import("./NormalModule")***REMOVED*** NormalModule */
/** @typedef {import("./RuntimeTemplate")***REMOVED*** RuntimeTemplate */
/** @typedef {import("./javascript/JavascriptParser")***REMOVED*** JavascriptParser */
/** @typedef {import("./javascript/JavascriptParser").DestructuringAssignmentProperty***REMOVED*** DestructuringAssignmentProperty */
/** @typedef {import("./javascript/JavascriptParser").Range***REMOVED*** Range */
/** @typedef {import("./logging/Logger").Logger***REMOVED*** Logger */

/** @typedef {null | undefined | RegExp | EXPECTED_FUNCTION | string | number | boolean | bigint | undefined***REMOVED*** CodeValuePrimitive */
/** @typedef {RecursiveArrayOrRecord<CodeValuePrimitive | RuntimeValue>***REMOVED*** CodeValue */

/**
 * @typedef {object***REMOVED*** RuntimeValueOptions
 * @property {string[]=***REMOVED*** fileDependencies
 * @property {string[]=***REMOVED*** contextDependencies
 * @property {string[]=***REMOVED*** missingDependencies
 * @property {string[]=***REMOVED*** buildDependencies
 * @property {string| (() => string)=***REMOVED*** version
 */

/** @typedef {string | Set<string>***REMOVED*** ValueCacheVersion */
/** @typedef {(value: { module: NormalModule, key: string, readonly version: ValueCacheVersion ***REMOVED***) => CodeValuePrimitive***REMOVED*** GeneratorFn */

class RuntimeValue {
	/**
	 * @param {GeneratorFn***REMOVED*** fn generator function
	 * @param {true | string[] | RuntimeValueOptions=***REMOVED*** options options
	 */
	constructor(fn, options) {
		this.fn = fn;
		if (Array.isArray(options)) {
			options = {
				fileDependencies: options
			***REMOVED***;
		***REMOVED***
		this.options = options || {***REMOVED***;
	***REMOVED***

	get fileDependencies() {
		return this.options === true ? true : this.options.fileDependencies;
	***REMOVED***

	/**
	 * @param {JavascriptParser***REMOVED*** parser the parser
	 * @param {ValueCacheVersions***REMOVED*** valueCacheVersions valueCacheVersions
	 * @param {string***REMOVED*** key the defined key
	 * @returns {CodeValuePrimitive***REMOVED*** code
	 */
	exec(parser, valueCacheVersions, key) {
		const buildInfo = /** @type {BuildInfo***REMOVED*** */ (parser.state.module.buildInfo);
		if (this.options === true) {
			buildInfo.cacheable = false;
		***REMOVED*** else {
			if (this.options.fileDependencies) {
				for (const dep of this.options.fileDependencies) {
					/** @type {NonNullable<BuildInfo["fileDependencies"]>***REMOVED*** */
					(buildInfo.fileDependencies).add(dep);
				***REMOVED***
			***REMOVED***
			if (this.options.contextDependencies) {
				for (const dep of this.options.contextDependencies) {
					/** @type {NonNullable<BuildInfo["contextDependencies"]>***REMOVED*** */
					(buildInfo.contextDependencies).add(dep);
				***REMOVED***
			***REMOVED***
			if (this.options.missingDependencies) {
				for (const dep of this.options.missingDependencies) {
					/** @type {NonNullable<BuildInfo["missingDependencies"]>***REMOVED*** */
					(buildInfo.missingDependencies).add(dep);
				***REMOVED***
			***REMOVED***
			if (this.options.buildDependencies) {
				for (const dep of this.options.buildDependencies) {
					/** @type {NonNullable<BuildInfo["buildDependencies"]>***REMOVED*** */
					(buildInfo.buildDependencies).add(dep);
				***REMOVED***
			***REMOVED***
		***REMOVED***

		return this.fn({
			module: parser.state.module,
			key,
			get version() {
				return /** @type {ValueCacheVersion***REMOVED*** */ (
					valueCacheVersions.get(VALUE_DEP_PREFIX + key)
				);
			***REMOVED***
		***REMOVED***);
	***REMOVED***

	getCacheVersion() {
		return this.options === true
			? undefined
			: (typeof this.options.version === "function"
					? this.options.version()
					: this.options.version) || "unset";
	***REMOVED***
***REMOVED***

/**
 * @param {Set<DestructuringAssignmentProperty> | undefined***REMOVED*** properties properties
 * @returns {Set<string> | undefined***REMOVED*** used keys
 */
function getObjKeys(properties) {
	if (!properties) return;
	return new Set([...properties].map(p => p.id));
***REMOVED***

/** @typedef {Set<string> | null***REMOVED*** ObjKeys */
/** @typedef {boolean | undefined | null***REMOVED*** AsiSafe */

/**
 * @param {EXPECTED_ANY[] | {[k: string]: EXPECTED_ANY***REMOVED******REMOVED*** obj obj
 * @param {JavascriptParser***REMOVED*** parser Parser
 * @param {ValueCacheVersions***REMOVED*** valueCacheVersions valueCacheVersions
 * @param {string***REMOVED*** key the defined key
 * @param {RuntimeTemplate***REMOVED*** runtimeTemplate the runtime template
 * @param {Logger***REMOVED*** logger the logger object
 * @param {AsiSafe=***REMOVED*** asiSafe asi safe (undefined: unknown, null: unneeded)
 * @param {ObjKeys=***REMOVED*** objKeys used keys
 * @returns {string***REMOVED*** code converted to string that evaluates
 */
const stringifyObj = (
	obj,
	parser,
	valueCacheVersions,
	key,
	runtimeTemplate,
	logger,
	asiSafe,
	objKeys
) => {
	let code;
	const arr = Array.isArray(obj);
	if (arr) {
		code = `[${obj
			.map(code =>
				toCode(
					code,
					parser,
					valueCacheVersions,
					key,
					runtimeTemplate,
					logger,
					null
				)
			)
			.join(",")***REMOVED***]`;
	***REMOVED*** else {
		let keys = Object.keys(obj);
		if (objKeys) {
			keys = objKeys.size === 0 ? [] : keys.filter(k => objKeys.has(k));
		***REMOVED***
		code = `{${keys
			.map(key => {
				const code = obj[key];
				return `${JSON.stringify(key)***REMOVED***:${toCode(
					code,
					parser,
					valueCacheVersions,
					key,
					runtimeTemplate,
					logger,
					null
				)***REMOVED***`;
			***REMOVED***)
			.join(",")***REMOVED******REMOVED***`;
	***REMOVED***

	switch (asiSafe) {
		case null:
			return code;
		case true:
			return arr ? code : `(${code***REMOVED***)`;
		case false:
			return arr ? `;${code***REMOVED***` : `;(${code***REMOVED***)`;
		default:
			return `/*#__PURE__*/Object(${code***REMOVED***)`;
	***REMOVED***
***REMOVED***;

/**
 * Convert code to a string that evaluates
 * @param {CodeValue***REMOVED*** code Code to evaluate
 * @param {JavascriptParser***REMOVED*** parser Parser
 * @param {ValueCacheVersions***REMOVED*** valueCacheVersions valueCacheVersions
 * @param {string***REMOVED*** key the defined key
 * @param {RuntimeTemplate***REMOVED*** runtimeTemplate the runtime template
 * @param {Logger***REMOVED*** logger the logger object
 * @param {boolean | undefined | null=***REMOVED*** asiSafe asi safe (undefined: unknown, null: unneeded)
 * @param {ObjKeys=***REMOVED*** objKeys used keys
 * @returns {string***REMOVED*** code converted to string that evaluates
 */
const toCode = (
	code,
	parser,
	valueCacheVersions,
	key,
	runtimeTemplate,
	logger,
	asiSafe,
	objKeys
) => {
	const transformToCode = () => {
		if (code === null) {
			return "null";
		***REMOVED***
		if (code === undefined) {
			return "undefined";
		***REMOVED***
		if (Object.is(code, -0)) {
			return "-0";
		***REMOVED***
		if (code instanceof RuntimeValue) {
			return toCode(
				code.exec(parser, valueCacheVersions, key),
				parser,
				valueCacheVersions,
				key,
				runtimeTemplate,
				logger,
				asiSafe
			);
		***REMOVED***
		if (code instanceof RegExp && code.toString) {
			return code.toString();
		***REMOVED***
		if (typeof code === "function" && code.toString) {
			return `(${code.toString()***REMOVED***)`;
		***REMOVED***
		if (typeof code === "object") {
			return stringifyObj(
				code,
				parser,
				valueCacheVersions,
				key,
				runtimeTemplate,
				logger,
				asiSafe,
				objKeys
			);
		***REMOVED***
		if (typeof code === "bigint") {
			return runtimeTemplate.supportsBigIntLiteral()
				? `${code***REMOVED***n`
				: `BigInt("${code***REMOVED***")`;
		***REMOVED***
		return `${code***REMOVED***`;
	***REMOVED***;

	const strCode = transformToCode();

	logger.debug(`Replaced "${key***REMOVED***" with "${strCode***REMOVED***"`);

	return strCode;
***REMOVED***;

/**
 * @param {CodeValue***REMOVED*** code code
 * @returns {string | undefined***REMOVED*** result
 */
const toCacheVersion = code => {
	if (code === null) {
		return "null";
	***REMOVED***
	if (code === undefined) {
		return "undefined";
	***REMOVED***
	if (Object.is(code, -0)) {
		return "-0";
	***REMOVED***
	if (code instanceof RuntimeValue) {
		return code.getCacheVersion();
	***REMOVED***
	if (code instanceof RegExp && code.toString) {
		return code.toString();
	***REMOVED***
	if (typeof code === "function" && code.toString) {
		return `(${code.toString()***REMOVED***)`;
	***REMOVED***
	if (typeof code === "object") {
		const items = Object.keys(code).map(key => ({
			key,
			value: toCacheVersion(
				/** @type {Record<string, EXPECTED_ANY>***REMOVED*** */
				(code)[key]
			)
		***REMOVED***));
		if (items.some(({ value ***REMOVED***) => value === undefined)) return;
		return `{${items.map(({ key, value ***REMOVED***) => `${key***REMOVED***: ${value***REMOVED***`).join(", ")***REMOVED******REMOVED***`;
	***REMOVED***
	if (typeof code === "bigint") {
		return `${code***REMOVED***n`;
	***REMOVED***
	return `${code***REMOVED***`;
***REMOVED***;

const PLUGIN_NAME = "DefinePlugin";
const VALUE_DEP_PREFIX = `webpack/${PLUGIN_NAME***REMOVED*** `;
const VALUE_DEP_MAIN = `webpack/${PLUGIN_NAME***REMOVED***_hash`;
const TYPEOF_OPERATOR_REGEXP = /^typeof\s+/;
const WEBPACK_REQUIRE_FUNCTION_REGEXP = new RegExp(
	`${RuntimeGlobals.require***REMOVED***\\s*(!?\\.)`
);
const WEBPACK_REQUIRE_IDENTIFIER_REGEXP = new RegExp(RuntimeGlobals.require);

class DefinePlugin {
	/**
	 * Create a new define plugin
	 * @param {Record<string, CodeValue>***REMOVED*** definitions A map of global object definitions
	 */
	constructor(definitions) {
		this.definitions = definitions;
	***REMOVED***

	/**
	 * @param {GeneratorFn***REMOVED*** fn generator function
	 * @param {true | string[] | RuntimeValueOptions=***REMOVED*** options options
	 * @returns {RuntimeValue***REMOVED*** runtime value
	 */
	static runtimeValue(fn, options) {
		return new RuntimeValue(fn, options);
	***REMOVED***

	/**
	 * Apply the plugin
	 * @param {Compiler***REMOVED*** compiler the compiler instance
	 * @returns {void***REMOVED***
	 */
	apply(compiler) {
		const definitions = this.definitions;
		compiler.hooks.compilation.tap(
			PLUGIN_NAME,
			(compilation, { normalModuleFactory ***REMOVED***) => {
				const logger = compilation.getLogger("webpack.DefinePlugin");
				compilation.dependencyTemplates.set(
					ConstDependency,
					new ConstDependency.Template()
				);
				const { runtimeTemplate ***REMOVED*** = compilation;

				const mainHash = createHash(
					/** @type {HashFunction***REMOVED*** */
					(compilation.outputOptions.hashFunction)
				);
				mainHash.update(
					/** @type {string***REMOVED*** */
					(compilation.valueCacheVersions.get(VALUE_DEP_MAIN)) || ""
				);

				/**
				 * Handler
				 * @param {JavascriptParser***REMOVED*** parser Parser
				 * @returns {void***REMOVED***
				 */
				const handler = parser => {
					const mainValue =
						/** @type {ValueCacheVersion***REMOVED*** */
						(compilation.valueCacheVersions.get(VALUE_DEP_MAIN));
					parser.hooks.program.tap(PLUGIN_NAME, () => {
						const buildInfo = /** @type {BuildInfo***REMOVED*** */ (
							parser.state.module.buildInfo
						);
						if (!buildInfo.valueDependencies)
							buildInfo.valueDependencies = new Map();
						buildInfo.valueDependencies.set(VALUE_DEP_MAIN, mainValue);
					***REMOVED***);

					/**
					 * @param {string***REMOVED*** key key
					 */
					const addValueDependency = key => {
						const buildInfo =
							/** @type {BuildInfo***REMOVED*** */
							(parser.state.module.buildInfo);
						/** @type {NonNullable<BuildInfo["valueDependencies"]>***REMOVED*** */
						(buildInfo.valueDependencies).set(
							VALUE_DEP_PREFIX + key,
							/** @type {ValueCacheVersion***REMOVED*** */
							(compilation.valueCacheVersions.get(VALUE_DEP_PREFIX + key))
						);
					***REMOVED***;

					/**
					 * @template T
					 * @param {string***REMOVED*** key key
					 * @param {(expression: Expression) => T***REMOVED*** fn fn
					 * @returns {(expression: Expression) => T***REMOVED*** result
					 */
					const withValueDependency =
						(key, fn) =>
						(...args) => {
							addValueDependency(key);
							return fn(...args);
						***REMOVED***;

					/**
					 * Walk definitions
					 * @param {Record<string, CodeValue>***REMOVED*** definitions Definitions map
					 * @param {string***REMOVED*** prefix Prefix string
					 * @returns {void***REMOVED***
					 */
					const walkDefinitions = (definitions, prefix) => {
						for (const key of Object.keys(definitions)) {
							const code = definitions[key];
							if (
								code &&
								typeof code === "object" &&
								!(code instanceof RuntimeValue) &&
								!(code instanceof RegExp)
							) {
								walkDefinitions(
									/** @type {Record<string, CodeValue>***REMOVED*** */ (code),
									`${prefix + key***REMOVED***.`
								);
								applyObjectDefine(prefix + key, code);
								continue;
							***REMOVED***
							applyDefineKey(prefix, key);
							applyDefine(prefix + key, code);
						***REMOVED***
					***REMOVED***;

					/**
					 * Apply define key
					 * @param {string***REMOVED*** prefix Prefix
					 * @param {string***REMOVED*** key Key
					 * @returns {void***REMOVED***
					 */
					const applyDefineKey = (prefix, key) => {
						const splittedKey = key.split(".");
						const firstKey = splittedKey[0];
						for (const [i, _] of splittedKey.slice(1).entries()) {
							const fullKey = prefix + splittedKey.slice(0, i + 1).join(".");
							parser.hooks.canRename.for(fullKey).tap(PLUGIN_NAME, () => {
								addValueDependency(key);
								if (
									parser.scope.definitions.get(firstKey) instanceof VariableInfo
								) {
									return false;
								***REMOVED***
								return true;
							***REMOVED***);
						***REMOVED***
					***REMOVED***;

					/**
					 * Apply Code
					 * @param {string***REMOVED*** key Key
					 * @param {CodeValue***REMOVED*** code Code
					 * @returns {void***REMOVED***
					 */
					const applyDefine = (key, code) => {
						const originalKey = key;
						const isTypeof = TYPEOF_OPERATOR_REGEXP.test(key);
						if (isTypeof) key = key.replace(TYPEOF_OPERATOR_REGEXP, "");
						let recurse = false;
						let recurseTypeof = false;
						if (!isTypeof) {
							parser.hooks.canRename.for(key).tap(PLUGIN_NAME, () => {
								addValueDependency(originalKey);
								return true;
							***REMOVED***);
							parser.hooks.evaluateIdentifier
								.for(key)
								.tap(PLUGIN_NAME, expr => {
									/**
									 * this is needed in case there is a recursion in the DefinePlugin
									 * to prevent an endless recursion
									 * e.g.: new DefinePlugin({
									 * "a": "b",
									 * "b": "a"
									 * ***REMOVED***);
									 */
									if (recurse) return;
									addValueDependency(originalKey);
									recurse = true;
									const res = parser.evaluate(
										toCode(
											code,
											parser,
											compilation.valueCacheVersions,
											key,
											runtimeTemplate,
											logger,
											null
										)
									);
									recurse = false;
									res.setRange(/** @type {Range***REMOVED*** */ (expr.range));
									return res;
								***REMOVED***);
							parser.hooks.expression.for(key).tap(PLUGIN_NAME, expr => {
								addValueDependency(originalKey);
								let strCode = toCode(
									code,
									parser,
									compilation.valueCacheVersions,
									originalKey,
									runtimeTemplate,
									logger,
									!parser.isAsiPosition(/** @type {Range***REMOVED*** */ (expr.range)[0]),
									null
								);

								if (parser.scope.inShorthand) {
									strCode = `${parser.scope.inShorthand***REMOVED***:${strCode***REMOVED***`;
								***REMOVED***

								if (WEBPACK_REQUIRE_FUNCTION_REGEXP.test(strCode)) {
									return toConstantDependency(parser, strCode, [
										RuntimeGlobals.require
									])(expr);
								***REMOVED*** else if (WEBPACK_REQUIRE_IDENTIFIER_REGEXP.test(strCode)) {
									return toConstantDependency(parser, strCode, [
										RuntimeGlobals.requireScope
									])(expr);
								***REMOVED***
								return toConstantDependency(parser, strCode)(expr);
							***REMOVED***);
						***REMOVED***
						parser.hooks.evaluateTypeof.for(key).tap(PLUGIN_NAME, expr => {
							/**
							 * this is needed in case there is a recursion in the DefinePlugin
							 * to prevent an endless recursion
							 * e.g.: new DefinePlugin({
							 * "typeof a": "typeof b",
							 * "typeof b": "typeof a"
							 * ***REMOVED***);
							 */
							if (recurseTypeof) return;
							recurseTypeof = true;
							addValueDependency(originalKey);
							const codeCode = toCode(
								code,
								parser,
								compilation.valueCacheVersions,
								originalKey,
								runtimeTemplate,
								logger,
								null
							);
							const typeofCode = isTypeof ? codeCode : `typeof (${codeCode***REMOVED***)`;
							const res = parser.evaluate(typeofCode);
							recurseTypeof = false;
							res.setRange(/** @type {Range***REMOVED*** */ (expr.range));
							return res;
						***REMOVED***);
						parser.hooks.typeof.for(key).tap(PLUGIN_NAME, expr => {
							addValueDependency(originalKey);
							const codeCode = toCode(
								code,
								parser,
								compilation.valueCacheVersions,
								originalKey,
								runtimeTemplate,
								logger,
								null
							);
							const typeofCode = isTypeof ? codeCode : `typeof (${codeCode***REMOVED***)`;
							const res = parser.evaluate(typeofCode);
							if (!res.isString()) return;
							return toConstantDependency(
								parser,
								JSON.stringify(res.string)
							).bind(parser)(expr);
						***REMOVED***);
					***REMOVED***;

					/**
					 * Apply Object
					 * @param {string***REMOVED*** key Key
					 * @param {object***REMOVED*** obj Object
					 * @returns {void***REMOVED***
					 */
					const applyObjectDefine = (key, obj) => {
						parser.hooks.canRename.for(key).tap(PLUGIN_NAME, () => {
							addValueDependency(key);
							return true;
						***REMOVED***);
						parser.hooks.evaluateIdentifier.for(key).tap(PLUGIN_NAME, expr => {
							addValueDependency(key);
							return new BasicEvaluatedExpression()
								.setTruthy()
								.setSideEffects(false)
								.setRange(/** @type {Range***REMOVED*** */ (expr.range));
						***REMOVED***);
						parser.hooks.evaluateTypeof
							.for(key)
							.tap(
								PLUGIN_NAME,
								withValueDependency(key, evaluateToString("object"))
							);
						parser.hooks.expression.for(key).tap(PLUGIN_NAME, expr => {
							addValueDependency(key);
							let strCode = stringifyObj(
								obj,
								parser,
								compilation.valueCacheVersions,
								key,
								runtimeTemplate,
								logger,
								!parser.isAsiPosition(/** @type {Range***REMOVED*** */ (expr.range)[0]),
								getObjKeys(parser.destructuringAssignmentPropertiesFor(expr))
							);

							if (parser.scope.inShorthand) {
								strCode = `${parser.scope.inShorthand***REMOVED***:${strCode***REMOVED***`;
							***REMOVED***

							if (WEBPACK_REQUIRE_FUNCTION_REGEXP.test(strCode)) {
								return toConstantDependency(parser, strCode, [
									RuntimeGlobals.require
								])(expr);
							***REMOVED*** else if (WEBPACK_REQUIRE_IDENTIFIER_REGEXP.test(strCode)) {
								return toConstantDependency(parser, strCode, [
									RuntimeGlobals.requireScope
								])(expr);
							***REMOVED***
							return toConstantDependency(parser, strCode)(expr);
						***REMOVED***);
						parser.hooks.typeof
							.for(key)
							.tap(
								PLUGIN_NAME,
								withValueDependency(
									key,
									toConstantDependency(parser, JSON.stringify("object"))
								)
							);
					***REMOVED***;

					walkDefinitions(definitions, "");
				***REMOVED***;

				normalModuleFactory.hooks.parser
					.for(JAVASCRIPT_MODULE_TYPE_AUTO)
					.tap(PLUGIN_NAME, handler);
				normalModuleFactory.hooks.parser
					.for(JAVASCRIPT_MODULE_TYPE_DYNAMIC)
					.tap(PLUGIN_NAME, handler);
				normalModuleFactory.hooks.parser
					.for(JAVASCRIPT_MODULE_TYPE_ESM)
					.tap(PLUGIN_NAME, handler);

				/**
				 * Walk definitions
				 * @param {Record<string, CodeValue>***REMOVED*** definitions Definitions map
				 * @param {string***REMOVED*** prefix Prefix string
				 * @returns {void***REMOVED***
				 */
				const walkDefinitionsForValues = (definitions, prefix) => {
					for (const key of Object.keys(definitions)) {
						const code = definitions[key];
						const version = /** @type {string***REMOVED*** */ (toCacheVersion(code));
						const name = VALUE_DEP_PREFIX + prefix + key;
						mainHash.update(`|${prefix***REMOVED***${key***REMOVED***`);
						const oldVersion = compilation.valueCacheVersions.get(name);
						if (oldVersion === undefined) {
							compilation.valueCacheVersions.set(name, version);
						***REMOVED*** else if (oldVersion !== version) {
							const warning = new WebpackError(
								`${PLUGIN_NAME***REMOVED***\nConflicting values for '${prefix + key***REMOVED***'`
							);
							warning.details = `'${oldVersion***REMOVED***' !== '${version***REMOVED***'`;
							warning.hideStack = true;
							compilation.warnings.push(warning);
						***REMOVED***
						if (
							code &&
							typeof code === "object" &&
							!(code instanceof RuntimeValue) &&
							!(code instanceof RegExp)
						) {
							walkDefinitionsForValues(
								/** @type {Record<string, CodeValue>***REMOVED*** */ (code),
								`${prefix + key***REMOVED***.`
							);
						***REMOVED***
					***REMOVED***
				***REMOVED***;

				walkDefinitionsForValues(definitions, "");

				compilation.valueCacheVersions.set(
					VALUE_DEP_MAIN,
					/** @type {string***REMOVED*** */ (mainHash.digest("hex").slice(0, 8))
				);
			***REMOVED***
		);
	***REMOVED***
***REMOVED***
module.exports = DefinePlugin;
