// webpack/lib/DllPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const DllEntryPlugin = require("./DllEntryPlugin");
const FlagAllModulesAsUsedPlugin = require("./FlagAllModulesAsUsedPlugin");
const LibManifestPlugin = require("./LibManifestPlugin");
const createSchemaValidation = require("./util/create-schema-validation");

/** @typedef {import("../declarations/plugins/DllPlugin").DllPluginOptions***REMOVED*** DllPluginOptions */
/** @typedef {import("./Compiler")***REMOVED*** Compiler */
/** @typedef {import("./DllEntryPlugin").Entries***REMOVED*** Entries */
/** @typedef {import("./DllEntryPlugin").Options***REMOVED*** Options */

const validate = createSchemaValidation(
	require("../schemas/plugins/DllPlugin.check.js"),
	() => require("../schemas/plugins/DllPlugin.json"),
	{
		name: "Dll Plugin",
		baseDataPath: "options"
	***REMOVED***
);

const PLUGIN_NAME = "DllPlugin";

class DllPlugin {
	/**
	 * @param {DllPluginOptions***REMOVED*** options options object
	 */
	constructor(options) {
		validate(options);
		this.options = {
			...options,
			entryOnly: options.entryOnly !== false
		***REMOVED***;
	***REMOVED***

	/**
	 * Apply the plugin
	 * @param {Compiler***REMOVED*** compiler the compiler instance
	 * @returns {void***REMOVED***
	 */
	apply(compiler) {
		compiler.hooks.entryOption.tap(PLUGIN_NAME, (context, entry) => {
			if (typeof entry !== "function") {
				for (const name of Object.keys(entry)) {
					/** @type {Options***REMOVED*** */
					const options = { name, filename: entry.filename ***REMOVED***;
					new DllEntryPlugin(
						context,
						/** @type {Entries***REMOVED*** */ (entry[name].import),
						options
					).apply(compiler);
				***REMOVED***
			***REMOVED*** else {
				throw new Error(
					`${PLUGIN_NAME***REMOVED*** doesn't support dynamic entry (function) yet`
				);
			***REMOVED***
			return true;
		***REMOVED***);
		new LibManifestPlugin(this.options).apply(compiler);
		if (!this.options.entryOnly) {
			new FlagAllModulesAsUsedPlugin(PLUGIN_NAME).apply(compiler);
		***REMOVED***
	***REMOVED***
***REMOVED***

module.exports = DllPlugin;
