// webpack/lib/Entrypoint.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const ChunkGroup = require("./ChunkGroup");

/** @typedef {import("../declarations/WebpackOptions").EntryDescriptionNormalized***REMOVED*** EntryDescription */
/** @typedef {import("./Chunk")***REMOVED*** Chunk */

/** @typedef {{ name?: string ***REMOVED*** & Omit<EntryDescription, "import">***REMOVED*** EntryOptions */

/**
 * Entrypoint serves as an encapsulation primitive for chunks that are
 * a part of a single ChunkGroup. They represent all bundles that need to be loaded for a
 * single instance of a page. Multi-page application architectures will typically yield multiple Entrypoint objects
 * inside of the compilation, whereas a Single Page App may only contain one with many lazy-loaded chunks.
 */
class Entrypoint extends ChunkGroup {
	/**
	 * Creates an instance of Entrypoint.
	 * @param {EntryOptions | string***REMOVED*** entryOptions the options for the entrypoint (or name)
	 * @param {boolean=***REMOVED*** initial false, when the entrypoint is not initial loaded
	 */
	constructor(entryOptions, initial = true) {
		if (typeof entryOptions === "string") {
			entryOptions = { name: entryOptions ***REMOVED***;
		***REMOVED***
		super({
			name: entryOptions.name
		***REMOVED***);
		this.options = entryOptions;
		/** @type {Chunk=***REMOVED*** */
		this._runtimeChunk = undefined;
		/** @type {Chunk=***REMOVED*** */
		this._entrypointChunk = undefined;
		/** @type {boolean***REMOVED*** */
		this._initial = initial;
	***REMOVED***

	/**
	 * @returns {boolean***REMOVED*** true, when this chunk group will be loaded on initial page load
	 */
	isInitial() {
		return this._initial;
	***REMOVED***

	/**
	 * Sets the runtimeChunk for an entrypoint.
	 * @param {Chunk***REMOVED*** chunk the chunk being set as the runtime chunk.
	 * @returns {void***REMOVED***
	 */
	setRuntimeChunk(chunk) {
		this._runtimeChunk = chunk;
	***REMOVED***

	/**
	 * Fetches the chunk reference containing the webpack bootstrap code
	 * @returns {Chunk | null***REMOVED*** returns the runtime chunk or null if there is none
	 */
	getRuntimeChunk() {
		if (this._runtimeChunk) return this._runtimeChunk;
		for (const parent of this.parentsIterable) {
			if (parent instanceof Entrypoint) return parent.getRuntimeChunk();
		***REMOVED***
		return null;
	***REMOVED***

	/**
	 * Sets the chunk with the entrypoint modules for an entrypoint.
	 * @param {Chunk***REMOVED*** chunk the chunk being set as the entrypoint chunk.
	 * @returns {void***REMOVED***
	 */
	setEntrypointChunk(chunk) {
		this._entrypointChunk = chunk;
	***REMOVED***

	/**
	 * Returns the chunk which contains the entrypoint modules
	 * (or at least the execution of them)
	 * @returns {Chunk***REMOVED*** chunk
	 */
	getEntrypointChunk() {
		return /** @type {Chunk***REMOVED*** */ (this._entrypointChunk);
	***REMOVED***

	/**
	 * @param {Chunk***REMOVED*** oldChunk chunk to be replaced
	 * @param {Chunk***REMOVED*** newChunk New chunk that will be replaced with
	 * @returns {boolean | undefined***REMOVED*** returns true if the replacement was successful
	 */
	replaceChunk(oldChunk, newChunk) {
		if (this._runtimeChunk === oldChunk) this._runtimeChunk = newChunk;
		if (this._entrypointChunk === oldChunk) this._entrypointChunk = newChunk;
		return super.replaceChunk(oldChunk, newChunk);
	***REMOVED***
***REMOVED***

module.exports = Entrypoint;
