// webpack/lib/HotModuleReplacementPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const { SyncBailHook ***REMOVED*** = require("tapable");
const { RawSource ***REMOVED*** = require("webpack-sources");
const ChunkGraph = require("./ChunkGraph");
const Compilation = require("./Compilation");
const HotUpdateChunk = require("./HotUpdateChunk");
const NormalModule = require("./NormalModule");
const RuntimeGlobals = require("./RuntimeGlobals");
const WebpackError = require("./WebpackError");
const ConstDependency = require("./dependencies/ConstDependency");
const ImportMetaHotAcceptDependency = require("./dependencies/ImportMetaHotAcceptDependency");
const ImportMetaHotDeclineDependency = require("./dependencies/ImportMetaHotDeclineDependency");
const ModuleHotAcceptDependency = require("./dependencies/ModuleHotAcceptDependency");
const ModuleHotDeclineDependency = require("./dependencies/ModuleHotDeclineDependency");
const HotModuleReplacementRuntimeModule = require("./hmr/HotModuleReplacementRuntimeModule");
const JavascriptParser = require("./javascript/JavascriptParser");
const {
	evaluateToIdentifier
***REMOVED*** = require("./javascript/JavascriptParserHelpers");
const { find, isSubset ***REMOVED*** = require("./util/SetHelpers");
const TupleSet = require("./util/TupleSet");
const { compareModulesById ***REMOVED*** = require("./util/comparators");
const {
	getRuntimeKey,
	keyToRuntime,
	forEachRuntime,
	mergeRuntimeOwned,
	subtractRuntime,
	intersectRuntime
***REMOVED*** = require("./util/runtime");

const {
	JAVASCRIPT_MODULE_TYPE_AUTO,
	JAVASCRIPT_MODULE_TYPE_DYNAMIC,
	JAVASCRIPT_MODULE_TYPE_ESM,
	WEBPACK_MODULE_TYPE_RUNTIME
***REMOVED*** = require("./ModuleTypeConstants");

/** @typedef {import("estree").CallExpression***REMOVED*** CallExpression */
/** @typedef {import("estree").Expression***REMOVED*** Expression */
/** @typedef {import("estree").SpreadElement***REMOVED*** SpreadElement */
/** @typedef {import("../declarations/WebpackOptions").OutputNormalized***REMOVED*** OutputNormalized */
/** @typedef {import("./Chunk")***REMOVED*** Chunk */
/** @typedef {import("./Chunk").ChunkId***REMOVED*** ChunkId */
/** @typedef {import("./ChunkGraph").ModuleId***REMOVED*** ModuleId */
/** @typedef {import("./Compilation").AssetInfo***REMOVED*** AssetInfo */
/** @typedef {import("./Compilation").Records***REMOVED*** Records */
/** @typedef {import("./Compiler")***REMOVED*** Compiler */
/** @typedef {import("./Dependency").DependencyLocation***REMOVED*** DependencyLocation */
/** @typedef {import("./Module")***REMOVED*** Module */
/** @typedef {import("./Module").BuildInfo***REMOVED*** BuildInfo */
/** @typedef {import("./RuntimeModule")***REMOVED*** RuntimeModule */
/** @typedef {import("./javascript/BasicEvaluatedExpression")***REMOVED*** BasicEvaluatedExpression */
/** @typedef {import("./javascript/JavascriptParserHelpers").Range***REMOVED*** Range */
/** @typedef {import("./util/runtime").RuntimeSpec***REMOVED*** RuntimeSpec */

/**
 * @typedef {object***REMOVED*** HMRJavascriptParserHooks
 * @property {SyncBailHook<[Expression | SpreadElement, string[]], void>***REMOVED*** hotAcceptCallback
 * @property {SyncBailHook<[CallExpression, string[]], void>***REMOVED*** hotAcceptWithoutCallback
 */

/** @typedef {{ updatedChunkIds: Set<ChunkId>, removedChunkIds: Set<ChunkId>, removedModules: Set<Module>, filename: string, assetInfo: AssetInfo ***REMOVED******REMOVED*** HotUpdateMainContentByRuntimeItem */
/** @typedef {Map<string, HotUpdateMainContentByRuntimeItem>***REMOVED*** HotUpdateMainContentByRuntime */

/** @type {WeakMap<JavascriptParser, HMRJavascriptParserHooks>***REMOVED*** */
const parserHooksMap = new WeakMap();

const PLUGIN_NAME = "HotModuleReplacementPlugin";

class HotModuleReplacementPlugin {
	/**
	 * @param {JavascriptParser***REMOVED*** parser the parser
	 * @returns {HMRJavascriptParserHooks***REMOVED*** the attached hooks
	 */
	static getParserHooks(parser) {
		if (!(parser instanceof JavascriptParser)) {
			throw new TypeError(
				"The 'parser' argument must be an instance of JavascriptParser"
			);
		***REMOVED***
		let hooks = parserHooksMap.get(parser);
		if (hooks === undefined) {
			hooks = {
				hotAcceptCallback: new SyncBailHook(["expression", "requests"]),
				hotAcceptWithoutCallback: new SyncBailHook(["expression", "requests"])
			***REMOVED***;
			parserHooksMap.set(parser, hooks);
		***REMOVED***
		return hooks;
	***REMOVED***

	/**
	 * Apply the plugin
	 * @param {Compiler***REMOVED*** compiler the compiler instance
	 * @returns {void***REMOVED***
	 */
	apply(compiler) {
		const { _backCompat: backCompat ***REMOVED*** = compiler;
		if (compiler.options.output.strictModuleErrorHandling === undefined)
			compiler.options.output.strictModuleErrorHandling = true;
		const runtimeRequirements = [RuntimeGlobals.module];

		/**
		 * @param {JavascriptParser***REMOVED*** parser the parser
		 * @param {typeof ModuleHotAcceptDependency***REMOVED*** ParamDependency dependency
		 * @returns {(expr: CallExpression) => boolean | undefined***REMOVED*** callback
		 */
		const createAcceptHandler = (parser, ParamDependency) => {
			const { hotAcceptCallback, hotAcceptWithoutCallback ***REMOVED*** =
				HotModuleReplacementPlugin.getParserHooks(parser);

			return expr => {
				const module = parser.state.module;
				const dep = new ConstDependency(
					`${module.moduleArgument***REMOVED***.hot.accept`,
					/** @type {Range***REMOVED*** */ (expr.callee.range),
					runtimeRequirements
				);
				dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
				module.addPresentationalDependency(dep);
				/** @type {BuildInfo***REMOVED*** */
				(module.buildInfo).moduleConcatenationBailout =
					"Hot Module Replacement";

				if (expr.arguments.length >= 1) {
					const arg = parser.evaluateExpression(expr.arguments[0]);
					/** @type {BasicEvaluatedExpression[]***REMOVED*** */
					let params = [];
					if (arg.isString()) {
						params = [arg];
					***REMOVED*** else if (arg.isArray()) {
						params =
							/** @type {BasicEvaluatedExpression[]***REMOVED*** */
							(arg.items).filter(param => param.isString());
					***REMOVED***
					/** @type {string[]***REMOVED*** */
					const requests = [];
					if (params.length > 0) {
						for (const [idx, param] of params.entries()) {
							const request = /** @type {string***REMOVED*** */ (param.string);
							const dep = new ParamDependency(
								request,
								/** @type {Range***REMOVED*** */ (param.range)
							);
							dep.optional = true;
							dep.loc = Object.create(
								/** @type {DependencyLocation***REMOVED*** */ (expr.loc)
							);
							dep.loc.index = idx;
							module.addDependency(dep);
							requests.push(request);
						***REMOVED***
						if (expr.arguments.length > 1) {
							hotAcceptCallback.call(expr.arguments[1], requests);
							for (let i = 1; i < expr.arguments.length; i++) {
								parser.walkExpression(expr.arguments[i]);
							***REMOVED***
							return true;
						***REMOVED***
						hotAcceptWithoutCallback.call(expr, requests);
						return true;
					***REMOVED***
				***REMOVED***
				parser.walkExpressions(expr.arguments);
				return true;
			***REMOVED***;
		***REMOVED***;

		/**
		 * @param {JavascriptParser***REMOVED*** parser the parser
		 * @param {typeof ModuleHotDeclineDependency***REMOVED*** ParamDependency dependency
		 * @returns {(expr: CallExpression) => boolean | undefined***REMOVED*** callback
		 */
		const createDeclineHandler = (parser, ParamDependency) => expr => {
			const module = parser.state.module;
			const dep = new ConstDependency(
				`${module.moduleArgument***REMOVED***.hot.decline`,
				/** @type {Range***REMOVED*** */ (expr.callee.range),
				runtimeRequirements
			);
			dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
			module.addPresentationalDependency(dep);
			/** @type {BuildInfo***REMOVED*** */
			(module.buildInfo).moduleConcatenationBailout = "Hot Module Replacement";
			if (expr.arguments.length === 1) {
				const arg = parser.evaluateExpression(expr.arguments[0]);
				/** @type {BasicEvaluatedExpression[]***REMOVED*** */
				let params = [];
				if (arg.isString()) {
					params = [arg];
				***REMOVED*** else if (arg.isArray()) {
					params =
						/** @type {BasicEvaluatedExpression[]***REMOVED*** */
						(arg.items).filter(param => param.isString());
				***REMOVED***
				for (const [idx, param] of params.entries()) {
					const dep = new ParamDependency(
						/** @type {string***REMOVED*** */ (param.string),
						/** @type {Range***REMOVED*** */ (param.range)
					);
					dep.optional = true;
					dep.loc = Object.create(/** @type {DependencyLocation***REMOVED*** */ (expr.loc));
					dep.loc.index = idx;
					module.addDependency(dep);
				***REMOVED***
			***REMOVED***
			return true;
		***REMOVED***;

		/**
		 * @param {JavascriptParser***REMOVED*** parser the parser
		 * @returns {(expr: Expression) => boolean | undefined***REMOVED*** callback
		 */
		const createHMRExpressionHandler = parser => expr => {
			const module = parser.state.module;
			const dep = new ConstDependency(
				`${module.moduleArgument***REMOVED***.hot`,
				/** @type {Range***REMOVED*** */ (expr.range),
				runtimeRequirements
			);
			dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
			module.addPresentationalDependency(dep);
			/** @type {BuildInfo***REMOVED*** */
			(module.buildInfo).moduleConcatenationBailout = "Hot Module Replacement";
			return true;
		***REMOVED***;

		/**
		 * @param {JavascriptParser***REMOVED*** parser the parser
		 * @returns {void***REMOVED***
		 */
		const applyModuleHot = parser => {
			parser.hooks.evaluateIdentifier.for("module.hot").tap(
				{
					name: PLUGIN_NAME,
					before: "NodeStuffPlugin"
				***REMOVED***,
				expr =>
					evaluateToIdentifier(
						"module.hot",
						"module",
						() => ["hot"],
						true
					)(expr)
			);
			parser.hooks.call
				.for("module.hot.accept")
				.tap(
					PLUGIN_NAME,
					createAcceptHandler(parser, ModuleHotAcceptDependency)
				);
			parser.hooks.call
				.for("module.hot.decline")
				.tap(
					PLUGIN_NAME,
					createDeclineHandler(parser, ModuleHotDeclineDependency)
				);
			parser.hooks.expression
				.for("module.hot")
				.tap(PLUGIN_NAME, createHMRExpressionHandler(parser));
		***REMOVED***;

		/**
		 * @param {JavascriptParser***REMOVED*** parser the parser
		 * @returns {void***REMOVED***
		 */
		const applyImportMetaHot = parser => {
			parser.hooks.evaluateIdentifier
				.for("import.meta.webpackHot")
				.tap(PLUGIN_NAME, expr =>
					evaluateToIdentifier(
						"import.meta.webpackHot",
						"import.meta",
						() => ["webpackHot"],
						true
					)(expr)
				);
			parser.hooks.call
				.for("import.meta.webpackHot.accept")
				.tap(
					PLUGIN_NAME,
					createAcceptHandler(parser, ImportMetaHotAcceptDependency)
				);
			parser.hooks.call
				.for("import.meta.webpackHot.decline")
				.tap(
					PLUGIN_NAME,
					createDeclineHandler(parser, ImportMetaHotDeclineDependency)
				);
			parser.hooks.expression
				.for("import.meta.webpackHot")
				.tap(PLUGIN_NAME, createHMRExpressionHandler(parser));
		***REMOVED***;

		compiler.hooks.compilation.tap(
			PLUGIN_NAME,
			(compilation, { normalModuleFactory ***REMOVED***) => {
				// This applies the HMR plugin only to the targeted compiler
				// It should not affect child compilations
				if (compilation.compiler !== compiler) return;

				// #region module.hot.* API
				compilation.dependencyFactories.set(
					ModuleHotAcceptDependency,
					normalModuleFactory
				);
				compilation.dependencyTemplates.set(
					ModuleHotAcceptDependency,
					new ModuleHotAcceptDependency.Template()
				);
				compilation.dependencyFactories.set(
					ModuleHotDeclineDependency,
					normalModuleFactory
				);
				compilation.dependencyTemplates.set(
					ModuleHotDeclineDependency,
					new ModuleHotDeclineDependency.Template()
				);
				// #endregion

				// #region import.meta.webpackHot.* API
				compilation.dependencyFactories.set(
					ImportMetaHotAcceptDependency,
					normalModuleFactory
				);
				compilation.dependencyTemplates.set(
					ImportMetaHotAcceptDependency,
					new ImportMetaHotAcceptDependency.Template()
				);
				compilation.dependencyFactories.set(
					ImportMetaHotDeclineDependency,
					normalModuleFactory
				);
				compilation.dependencyTemplates.set(
					ImportMetaHotDeclineDependency,
					new ImportMetaHotDeclineDependency.Template()
				);
				// #endregion

				let hotIndex = 0;
				/** @type {Record<string, string>***REMOVED*** */
				const fullHashChunkModuleHashes = {***REMOVED***;
				/** @type {Record<string, string>***REMOVED*** */
				const chunkModuleHashes = {***REMOVED***;

				compilation.hooks.record.tap(PLUGIN_NAME, (compilation, records) => {
					if (records.hash === compilation.hash) return;
					const chunkGraph = compilation.chunkGraph;
					records.hash = compilation.hash;
					records.hotIndex = hotIndex;
					records.fullHashChunkModuleHashes = fullHashChunkModuleHashes;
					records.chunkModuleHashes = chunkModuleHashes;
					records.chunkHashes = {***REMOVED***;
					records.chunkRuntime = {***REMOVED***;
					for (const chunk of compilation.chunks) {
						const chunkId = /** @type {ChunkId***REMOVED*** */ (chunk.id);
						records.chunkHashes[chunkId] = chunk.hash;
						records.chunkRuntime[chunkId] = getRuntimeKey(chunk.runtime);
					***REMOVED***
					records.chunkModuleIds = {***REMOVED***;
					for (const chunk of compilation.chunks) {
						records.chunkModuleIds[/** @type {ChunkId***REMOVED*** */ (chunk.id)] =
							Array.from(
								chunkGraph.getOrderedChunkModulesIterable(
									chunk,
									compareModulesById(chunkGraph)
								),
								m => chunkGraph.getModuleId(m)
							);
					***REMOVED***
				***REMOVED***);
				/** @type {TupleSet<Module, Chunk>***REMOVED*** */
				const updatedModules = new TupleSet();
				/** @type {TupleSet<Module, Chunk>***REMOVED*** */
				const fullHashModules = new TupleSet();
				/** @type {TupleSet<Module, RuntimeSpec>***REMOVED*** */
				const nonCodeGeneratedModules = new TupleSet();
				compilation.hooks.fullHash.tap(PLUGIN_NAME, hash => {
					const chunkGraph = compilation.chunkGraph;
					const records = /** @type {Records***REMOVED*** */ (compilation.records);
					for (const chunk of compilation.chunks) {
						/**
						 * @param {Module***REMOVED*** module module
						 * @returns {string***REMOVED*** module hash
						 */
						const getModuleHash = module => {
							if (
								compilation.codeGenerationResults.has(module, chunk.runtime)
							) {
								return compilation.codeGenerationResults.getHash(
									module,
									chunk.runtime
								);
							***REMOVED***
							nonCodeGeneratedModules.add(module, chunk.runtime);
							return chunkGraph.getModuleHash(module, chunk.runtime);
						***REMOVED***;
						const fullHashModulesInThisChunk =
							chunkGraph.getChunkFullHashModulesSet(chunk);
						if (fullHashModulesInThisChunk !== undefined) {
							for (const module of fullHashModulesInThisChunk) {
								fullHashModules.add(module, chunk);
							***REMOVED***
						***REMOVED***
						const modules = chunkGraph.getChunkModulesIterable(chunk);
						if (modules !== undefined) {
							if (records.chunkModuleHashes) {
								if (fullHashModulesInThisChunk !== undefined) {
									for (const module of modules) {
										const key = `${chunk.id***REMOVED***|${module.identifier()***REMOVED***`;
										const hash = getModuleHash(module);
										if (
											fullHashModulesInThisChunk.has(
												/** @type {RuntimeModule***REMOVED*** */ (module)
											)
										) {
											if (records.fullHashChunkModuleHashes[key] !== hash) {
												updatedModules.add(module, chunk);
											***REMOVED***
											fullHashChunkModuleHashes[key] = hash;
										***REMOVED*** else {
											if (records.chunkModuleHashes[key] !== hash) {
												updatedModules.add(module, chunk);
											***REMOVED***
											chunkModuleHashes[key] = hash;
										***REMOVED***
									***REMOVED***
								***REMOVED*** else {
									for (const module of modules) {
										const key = `${chunk.id***REMOVED***|${module.identifier()***REMOVED***`;
										const hash = getModuleHash(module);
										if (records.chunkModuleHashes[key] !== hash) {
											updatedModules.add(module, chunk);
										***REMOVED***
										chunkModuleHashes[key] = hash;
									***REMOVED***
								***REMOVED***
							***REMOVED*** else if (fullHashModulesInThisChunk !== undefined) {
								for (const module of modules) {
									const key = `${chunk.id***REMOVED***|${module.identifier()***REMOVED***`;
									const hash = getModuleHash(module);
									if (
										fullHashModulesInThisChunk.has(
											/** @type {RuntimeModule***REMOVED*** */ (module)
										)
									) {
										fullHashChunkModuleHashes[key] = hash;
									***REMOVED*** else {
										chunkModuleHashes[key] = hash;
									***REMOVED***
								***REMOVED***
							***REMOVED*** else {
								for (const module of modules) {
									const key = `${chunk.id***REMOVED***|${module.identifier()***REMOVED***`;
									const hash = getModuleHash(module);
									chunkModuleHashes[key] = hash;
								***REMOVED***
							***REMOVED***
						***REMOVED***
					***REMOVED***

					hotIndex = records.hotIndex || 0;
					if (updatedModules.size > 0) hotIndex++;

					hash.update(`${hotIndex***REMOVED***`);
				***REMOVED***);
				compilation.hooks.processAssets.tap(
					{
						name: PLUGIN_NAME,
						stage: Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL
					***REMOVED***,
					() => {
						const chunkGraph = compilation.chunkGraph;
						const records = /** @type {Records***REMOVED*** */ (compilation.records);
						if (records.hash === compilation.hash) return;
						if (
							!records.chunkModuleHashes ||
							!records.chunkHashes ||
							!records.chunkModuleIds
						) {
							return;
						***REMOVED***
						for (const [module, chunk] of fullHashModules) {
							const key = `${chunk.id***REMOVED***|${module.identifier()***REMOVED***`;
							const hash = nonCodeGeneratedModules.has(module, chunk.runtime)
								? chunkGraph.getModuleHash(module, chunk.runtime)
								: compilation.codeGenerationResults.getHash(
										module,
										chunk.runtime
									);
							if (records.chunkModuleHashes[key] !== hash) {
								updatedModules.add(module, chunk);
							***REMOVED***
							chunkModuleHashes[key] = hash;
						***REMOVED***

						/** @type {HotUpdateMainContentByRuntime***REMOVED*** */
						const hotUpdateMainContentByRuntime = new Map();
						let allOldRuntime;
						for (const key of Object.keys(records.chunkRuntime)) {
							const runtime = keyToRuntime(records.chunkRuntime[key]);
							allOldRuntime = mergeRuntimeOwned(allOldRuntime, runtime);
						***REMOVED***
						forEachRuntime(allOldRuntime, runtime => {
							const { path: filename, info: assetInfo ***REMOVED*** =
								compilation.getPathWithInfo(
									/** @type {NonNullable<OutputNormalized["hotUpdateMainFilename"]>***REMOVED*** */
									(compilation.outputOptions.hotUpdateMainFilename),
									{
										hash: records.hash,
										runtime
									***REMOVED***
								);
							hotUpdateMainContentByRuntime.set(
								/** @type {string***REMOVED*** */ (runtime),
								{
									updatedChunkIds: new Set(),
									removedChunkIds: new Set(),
									removedModules: new Set(),
									filename,
									assetInfo
								***REMOVED***
							);
						***REMOVED***);
						if (hotUpdateMainContentByRuntime.size === 0) return;

						// Create a list of all active modules to verify which modules are removed completely
						/** @type {Map<number|string, Module>***REMOVED*** */
						const allModules = new Map();
						for (const module of compilation.modules) {
							const id =
								/** @type {ModuleId***REMOVED*** */
								(chunkGraph.getModuleId(module));
							allModules.set(id, module);
						***REMOVED***

						// List of completely removed modules
						/** @type {Set<string | number>***REMOVED*** */
						const completelyRemovedModules = new Set();

						for (const key of Object.keys(records.chunkHashes)) {
							const oldRuntime = keyToRuntime(records.chunkRuntime[key]);
							/** @type {Module[]***REMOVED*** */
							const remainingModules = [];
							// Check which modules are removed
							for (const id of records.chunkModuleIds[key]) {
								const module = allModules.get(id);
								if (module === undefined) {
									completelyRemovedModules.add(id);
								***REMOVED*** else {
									remainingModules.push(module);
								***REMOVED***
							***REMOVED***

							/** @type {ChunkId | null***REMOVED*** */
							let chunkId;
							let newModules;
							let newRuntimeModules;
							let newFullHashModules;
							let newDependentHashModules;
							let newRuntime;
							let removedFromRuntime;
							const currentChunk = find(
								compilation.chunks,
								chunk => `${chunk.id***REMOVED***` === key
							);
							if (currentChunk) {
								chunkId = currentChunk.id;
								newRuntime = intersectRuntime(
									currentChunk.runtime,
									allOldRuntime
								);
								if (newRuntime === undefined) continue;
								newModules = chunkGraph
									.getChunkModules(currentChunk)
									.filter(module => updatedModules.has(module, currentChunk));
								newRuntimeModules = Array.from(
									chunkGraph.getChunkRuntimeModulesIterable(currentChunk)
								).filter(module => updatedModules.has(module, currentChunk));
								const fullHashModules =
									chunkGraph.getChunkFullHashModulesIterable(currentChunk);
								newFullHashModules =
									fullHashModules &&
									Array.from(fullHashModules).filter(module =>
										updatedModules.has(module, currentChunk)
									);
								const dependentHashModules =
									chunkGraph.getChunkDependentHashModulesIterable(currentChunk);
								newDependentHashModules =
									dependentHashModules &&
									Array.from(dependentHashModules).filter(module =>
										updatedModules.has(module, currentChunk)
									);
								removedFromRuntime = subtractRuntime(oldRuntime, newRuntime);
							***REMOVED*** else {
								// chunk has completely removed
								chunkId = `${Number(key)***REMOVED***` === key ? Number(key) : key;
								removedFromRuntime = oldRuntime;
								newRuntime = oldRuntime;
							***REMOVED***
							if (removedFromRuntime) {
								// chunk was removed from some runtimes
								forEachRuntime(removedFromRuntime, runtime => {
									const item =
										/** @type {HotUpdateMainContentByRuntimeItem***REMOVED*** */
										(
											hotUpdateMainContentByRuntime.get(
												/** @type {string***REMOVED*** */ (runtime)
											)
										);
									item.removedChunkIds.add(/** @type {ChunkId***REMOVED*** */ (chunkId));
								***REMOVED***);
								// dispose modules from the chunk in these runtimes
								// where they are no longer in this runtime
								for (const module of remainingModules) {
									const moduleKey = `${key***REMOVED***|${module.identifier()***REMOVED***`;
									const oldHash = records.chunkModuleHashes[moduleKey];
									const runtimes = chunkGraph.getModuleRuntimes(module);
									if (oldRuntime === newRuntime && runtimes.has(newRuntime)) {
										// Module is still in the same runtime combination
										const hash = nonCodeGeneratedModules.has(module, newRuntime)
											? chunkGraph.getModuleHash(module, newRuntime)
											: compilation.codeGenerationResults.getHash(
													module,
													newRuntime
												);
										if (hash !== oldHash) {
											if (module.type === WEBPACK_MODULE_TYPE_RUNTIME) {
												newRuntimeModules = newRuntimeModules || [];
												newRuntimeModules.push(
													/** @type {RuntimeModule***REMOVED*** */ (module)
												);
											***REMOVED*** else {
												newModules = newModules || [];
												newModules.push(module);
											***REMOVED***
										***REMOVED***
									***REMOVED*** else {
										// module is no longer in this runtime combination
										// We (incorrectly) assume that it's not in an overlapping runtime combination
										// and dispose it from the main runtimes the chunk was removed from
										forEachRuntime(removedFromRuntime, runtime => {
											// If the module is still used in this runtime, do not dispose it
											// This could create a bad runtime state where the module is still loaded,
											// but no chunk which contains it. This means we don't receive further HMR updates
											// to this module and that's bad.
											// TODO force load one of the chunks which contains the module
											for (const moduleRuntime of runtimes) {
												if (typeof moduleRuntime === "string") {
													if (moduleRuntime === runtime) return;
												***REMOVED*** else if (
													moduleRuntime !== undefined &&
													moduleRuntime.has(/** @type {string***REMOVED*** */ (runtime))
												)
													return;
											***REMOVED***
											const item =
												/** @type {HotUpdateMainContentByRuntimeItem***REMOVED*** */ (
													hotUpdateMainContentByRuntime.get(
														/** @type {string***REMOVED*** */ (runtime)
													)
												);
											item.removedModules.add(module);
										***REMOVED***);
									***REMOVED***
								***REMOVED***
							***REMOVED***
							if (
								(newModules && newModules.length > 0) ||
								(newRuntimeModules && newRuntimeModules.length > 0)
							) {
								const hotUpdateChunk = new HotUpdateChunk();
								if (backCompat)
									ChunkGraph.setChunkGraphForChunk(hotUpdateChunk, chunkGraph);
								hotUpdateChunk.id = chunkId;
								hotUpdateChunk.runtime = currentChunk
									? currentChunk.runtime
									: newRuntime;
								if (currentChunk) {
									for (const group of currentChunk.groupsIterable)
										hotUpdateChunk.addGroup(group);
								***REMOVED***
								chunkGraph.attachModules(hotUpdateChunk, newModules || []);
								chunkGraph.attachRuntimeModules(
									hotUpdateChunk,
									newRuntimeModules || []
								);
								if (newFullHashModules) {
									chunkGraph.attachFullHashModules(
										hotUpdateChunk,
										newFullHashModules
									);
								***REMOVED***
								if (newDependentHashModules) {
									chunkGraph.attachDependentHashModules(
										hotUpdateChunk,
										newDependentHashModules
									);
								***REMOVED***
								const renderManifest = compilation.getRenderManifest({
									chunk: hotUpdateChunk,
									hash: records.hash,
									fullHash: records.hash,
									outputOptions: compilation.outputOptions,
									moduleTemplates: compilation.moduleTemplates,
									dependencyTemplates: compilation.dependencyTemplates,
									codeGenerationResults: compilation.codeGenerationResults,
									runtimeTemplate: compilation.runtimeTemplate,
									moduleGraph: compilation.moduleGraph,
									chunkGraph
								***REMOVED***);
								for (const entry of renderManifest) {
									/** @type {string***REMOVED*** */
									let filename;
									/** @type {AssetInfo***REMOVED*** */
									let assetInfo;
									if ("filename" in entry) {
										filename = entry.filename;
										assetInfo = entry.info;
									***REMOVED*** else {
										({ path: filename, info: assetInfo ***REMOVED*** =
											compilation.getPathWithInfo(
												entry.filenameTemplate,
												entry.pathOptions
											));
									***REMOVED***
									const source = entry.render();
									compilation.additionalChunkAssets.push(filename);
									compilation.emitAsset(filename, source, {
										hotModuleReplacement: true,
										...assetInfo
									***REMOVED***);
									if (currentChunk) {
										currentChunk.files.add(filename);
										compilation.hooks.chunkAsset.call(currentChunk, filename);
									***REMOVED***
								***REMOVED***
								forEachRuntime(newRuntime, runtime => {
									const item =
										/** @type {HotUpdateMainContentByRuntimeItem***REMOVED*** */ (
											hotUpdateMainContentByRuntime.get(
												/** @type {string***REMOVED*** */ (runtime)
											)
										);
									item.updatedChunkIds.add(/** @type {ChunkId***REMOVED*** */ (chunkId));
								***REMOVED***);
							***REMOVED***
						***REMOVED***
						const completelyRemovedModulesArray = Array.from(
							completelyRemovedModules
						);
						const hotUpdateMainContentByFilename = new Map();
						for (const {
							removedChunkIds,
							removedModules,
							updatedChunkIds,
							filename,
							assetInfo
						***REMOVED*** of hotUpdateMainContentByRuntime.values()) {
							const old = hotUpdateMainContentByFilename.get(filename);
							if (
								old &&
								(!isSubset(old.removedChunkIds, removedChunkIds) ||
									!isSubset(old.removedModules, removedModules) ||
									!isSubset(old.updatedChunkIds, updatedChunkIds))
							) {
								compilation.warnings.push(
									new WebpackError(`HotModuleReplacementPlugin
The configured output.hotUpdateMainFilename doesn't lead to unique filenames per runtime and HMR update differs between runtimes.
This might lead to incorrect runtime behavior of the applied update.
To fix this, make sure to include [runtime] in the output.hotUpdateMainFilename option, or use the default config.`)
								);
								for (const chunkId of removedChunkIds)
									old.removedChunkIds.add(chunkId);
								for (const chunkId of removedModules)
									old.removedModules.add(chunkId);
								for (const chunkId of updatedChunkIds)
									old.updatedChunkIds.add(chunkId);
								continue;
							***REMOVED***
							hotUpdateMainContentByFilename.set(filename, {
								removedChunkIds,
								removedModules,
								updatedChunkIds,
								assetInfo
							***REMOVED***);
						***REMOVED***
						for (const [
							filename,
							{ removedChunkIds, removedModules, updatedChunkIds, assetInfo ***REMOVED***
						] of hotUpdateMainContentByFilename) {
							const hotUpdateMainJson = {
								c: Array.from(updatedChunkIds),
								r: Array.from(removedChunkIds),
								m:
									removedModules.size === 0
										? completelyRemovedModulesArray
										: completelyRemovedModulesArray.concat(
												Array.from(
													removedModules,
													m =>
														/** @type {ModuleId***REMOVED*** */ (chunkGraph.getModuleId(m))
												)
											)
							***REMOVED***;

							const source = new RawSource(JSON.stringify(hotUpdateMainJson));
							compilation.emitAsset(filename, source, {
								hotModuleReplacement: true,
								...assetInfo
							***REMOVED***);
						***REMOVED***
					***REMOVED***
				);

				compilation.hooks.additionalTreeRuntimeRequirements.tap(
					PLUGIN_NAME,
					(chunk, runtimeRequirements) => {
						runtimeRequirements.add(RuntimeGlobals.hmrDownloadManifest);
						runtimeRequirements.add(RuntimeGlobals.hmrDownloadUpdateHandlers);
						runtimeRequirements.add(RuntimeGlobals.interceptModuleExecution);
						runtimeRequirements.add(RuntimeGlobals.moduleCache);
						compilation.addRuntimeModule(
							chunk,
							new HotModuleReplacementRuntimeModule()
						);
					***REMOVED***
				);

				normalModuleFactory.hooks.parser
					.for(JAVASCRIPT_MODULE_TYPE_AUTO)
					.tap(PLUGIN_NAME, parser => {
						applyModuleHot(parser);
						applyImportMetaHot(parser);
					***REMOVED***);
				normalModuleFactory.hooks.parser
					.for(JAVASCRIPT_MODULE_TYPE_DYNAMIC)
					.tap(PLUGIN_NAME, parser => {
						applyModuleHot(parser);
					***REMOVED***);
				normalModuleFactory.hooks.parser
					.for(JAVASCRIPT_MODULE_TYPE_ESM)
					.tap(PLUGIN_NAME, parser => {
						applyImportMetaHot(parser);
					***REMOVED***);
				normalModuleFactory.hooks.module.tap(PLUGIN_NAME, module => {
					module.hot = true;
					return module;
				***REMOVED***);

				NormalModule.getCompilationHooks(compilation).loader.tap(
					PLUGIN_NAME,
					context => {
						context.hot = true;
					***REMOVED***
				);
			***REMOVED***
		);
	***REMOVED***
***REMOVED***

module.exports = HotModuleReplacementPlugin;
