// webpack/lib/IgnorePlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const RawModule = require("./RawModule");
const EntryDependency = require("./dependencies/EntryDependency");
const createSchemaValidation = require("./util/create-schema-validation");

/** @typedef {import("../declarations/plugins/IgnorePlugin").IgnorePluginOptions***REMOVED*** IgnorePluginOptions */
/** @typedef {import("./Compiler")***REMOVED*** Compiler */
/** @typedef {import("./NormalModuleFactory").ResolveData***REMOVED*** ResolveData */

const validate = createSchemaValidation(
	require("../schemas/plugins/IgnorePlugin.check.js"),
	() => require("../schemas/plugins/IgnorePlugin.json"),
	{
		name: "Ignore Plugin",
		baseDataPath: "options"
	***REMOVED***
);

const PLUGIN_NAME = "IgnorePlugin";

class IgnorePlugin {
	/**
	 * @param {IgnorePluginOptions***REMOVED*** options IgnorePlugin options
	 */
	constructor(options) {
		validate(options);
		this.options = options;
		this.checkIgnore = this.checkIgnore.bind(this);
	***REMOVED***

	/**
	 * Note that if "contextRegExp" is given, both the "resourceRegExp" and "contextRegExp" have to match.
	 * @param {ResolveData***REMOVED*** resolveData resolve data
	 * @returns {false|undefined***REMOVED*** returns false when the request should be ignored, otherwise undefined
	 */
	checkIgnore(resolveData) {
		if (
			"checkResource" in this.options &&
			this.options.checkResource &&
			this.options.checkResource(resolveData.request, resolveData.context)
		) {
			return false;
		***REMOVED***

		if (
			"resourceRegExp" in this.options &&
			this.options.resourceRegExp &&
			this.options.resourceRegExp.test(resolveData.request)
		) {
			if ("contextRegExp" in this.options && this.options.contextRegExp) {
				// if "contextRegExp" is given,
				// both the "resourceRegExp" and "contextRegExp" have to match.
				if (this.options.contextRegExp.test(resolveData.context)) {
					return false;
				***REMOVED***
			***REMOVED*** else {
				return false;
			***REMOVED***
		***REMOVED***
	***REMOVED***

	/**
	 * Apply the plugin
	 * @param {Compiler***REMOVED*** compiler the compiler instance
	 * @returns {void***REMOVED***
	 */
	apply(compiler) {
		compiler.hooks.normalModuleFactory.tap(PLUGIN_NAME, nmf => {
			nmf.hooks.beforeResolve.tap(PLUGIN_NAME, resolveData => {
				const result = this.checkIgnore(resolveData);

				if (
					result === false &&
					resolveData.dependencies.length > 0 &&
					resolveData.dependencies[0] instanceof EntryDependency
				) {
					resolveData.ignoredModule = new RawModule(
						"",
						"ignored-entry-module",
						"(ignored-entry-module)"
					);
				***REMOVED***

				return result;
			***REMOVED***);
		***REMOVED***);
		compiler.hooks.contextModuleFactory.tap(PLUGIN_NAME, cmf => {
			cmf.hooks.beforeResolve.tap(PLUGIN_NAME, this.checkIgnore);
		***REMOVED***);
	***REMOVED***
***REMOVED***

module.exports = IgnorePlugin;
