// webpack/lib/index.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const util = require("util");
const memoize = require("./util/memoize");

/** @typedef {import("../declarations/WebpackOptions").Entry***REMOVED*** Entry */
/** @typedef {import("../declarations/WebpackOptions").EntryNormalized***REMOVED*** EntryNormalized */
/** @typedef {import("../declarations/WebpackOptions").EntryObject***REMOVED*** EntryObject */
/** @typedef {import("../declarations/WebpackOptions").ExternalItem***REMOVED*** ExternalItem */
/** @typedef {import("../declarations/WebpackOptions").ExternalItemFunction***REMOVED*** ExternalItemFunction */
/** @typedef {import("../declarations/WebpackOptions").ExternalItemFunctionCallback***REMOVED*** ExternalItemFunctionCallback */
/** @typedef {import("../declarations/WebpackOptions").ExternalItemFunctionData***REMOVED*** ExternalItemFunctionData */
/** @typedef {import("../declarations/WebpackOptions").ExternalItemFunctionDataGetResolve***REMOVED*** ExternalItemFunctionDataGetResolve */
/** @typedef {import("../declarations/WebpackOptions").ExternalItemFunctionDataGetResolveCallbackResult***REMOVED*** ExternalItemFunctionDataGetResolveCallbackResult */
/** @typedef {import("../declarations/WebpackOptions").ExternalItemFunctionDataGetResolveResult***REMOVED*** ExternalItemFunctionDataGetResolveResult */
/** @typedef {import("../declarations/WebpackOptions").ExternalItemFunctionPromise***REMOVED*** ExternalItemFunctionPromise */
/** @typedef {import("../declarations/WebpackOptions").ExternalItemObjectKnown***REMOVED*** ExternalItemObjectKnown */
/** @typedef {import("../declarations/WebpackOptions").ExternalItemObjectUnknown***REMOVED*** ExternalItemObjectUnknown */
/** @typedef {import("../declarations/WebpackOptions").ExternalItemValue***REMOVED*** ExternalItemValue */
/** @typedef {import("../declarations/WebpackOptions").Externals***REMOVED*** Externals */
/** @typedef {import("../declarations/WebpackOptions").FileCacheOptions***REMOVED*** FileCacheOptions */
/** @typedef {import("../declarations/WebpackOptions").GeneratorOptionsByModuleTypeKnown***REMOVED*** GeneratorOptionsByModuleTypeKnown */
/** @typedef {import("../declarations/WebpackOptions").LibraryOptions***REMOVED*** LibraryOptions */
/** @typedef {import("../declarations/WebpackOptions").MemoryCacheOptions***REMOVED*** MemoryCacheOptions */
/** @typedef {import("../declarations/WebpackOptions").ModuleOptions***REMOVED*** ModuleOptions */
/** @typedef {import("../declarations/WebpackOptions").ParserOptionsByModuleTypeKnown***REMOVED*** ParserOptionsByModuleTypeKnown */
/** @typedef {import("../declarations/WebpackOptions").ResolveOptions***REMOVED*** ResolveOptions */
/** @typedef {import("../declarations/WebpackOptions").RuleSetCondition***REMOVED*** RuleSetCondition */
/** @typedef {import("../declarations/WebpackOptions").RuleSetConditionAbsolute***REMOVED*** RuleSetConditionAbsolute */
/** @typedef {import("../declarations/WebpackOptions").RuleSetRule***REMOVED*** RuleSetRule */
/** @typedef {import("../declarations/WebpackOptions").RuleSetUse***REMOVED*** RuleSetUse */
/** @typedef {import("../declarations/WebpackOptions").RuleSetUseFunction***REMOVED*** RuleSetUseFunction */
/** @typedef {import("../declarations/WebpackOptions").RuleSetUseItem***REMOVED*** RuleSetUseItem */
/** @typedef {import("../declarations/WebpackOptions").StatsOptions***REMOVED*** StatsOptions */
/** @typedef {import("../declarations/WebpackOptions").WebpackOptions***REMOVED*** Configuration */
/** @typedef {import("../declarations/WebpackOptions").WebpackOptionsNormalized***REMOVED*** WebpackOptionsNormalized */
/** @typedef {import("../declarations/WebpackOptions").WebpackPluginFunction***REMOVED*** WebpackPluginFunction */
/** @typedef {import("../declarations/WebpackOptions").WebpackPluginInstance***REMOVED*** WebpackPluginInstance */
/** @typedef {import("./ChunkGroup")***REMOVED*** ChunkGroup */
/** @typedef {import("./Compilation").Asset***REMOVED*** Asset */
/** @typedef {import("./Compilation").AssetInfo***REMOVED*** AssetInfo */
/** @typedef {import("./Compilation").EntryOptions***REMOVED*** EntryOptions */
/** @typedef {import("./Compilation").PathData***REMOVED*** PathData */
/** @typedef {import("./Compiler").AssetEmittedInfo***REMOVED*** AssetEmittedInfo */
/** @typedef {import("./Entrypoint")***REMOVED*** Entrypoint */
/** @typedef {import("./MultiCompiler").MultiCompilerOptions***REMOVED*** MultiCompilerOptions */
/** @typedef {import("./MultiStats")***REMOVED*** MultiStats */
/** @typedef {import("./NormalModuleFactory").ResolveData***REMOVED*** ResolveData */
/** @typedef {import("./Parser").ParserState***REMOVED*** ParserState */
/** @typedef {import("./ResolverFactory").ResolvePluginInstance***REMOVED*** ResolvePluginInstance */
/** @typedef {import("./ResolverFactory").Resolver***REMOVED*** Resolver */
/** @typedef {import("./Watching")***REMOVED*** Watching */
/** @typedef {import("./cli").Argument***REMOVED*** Argument */
/** @typedef {import("./cli").Problem***REMOVED*** Problem */
/** @typedef {import("./stats/DefaultStatsFactoryPlugin").StatsAsset***REMOVED*** StatsAsset */
/** @typedef {import("./stats/DefaultStatsFactoryPlugin").StatsChunk***REMOVED*** StatsChunk */
/** @typedef {import("./stats/DefaultStatsFactoryPlugin").StatsChunkGroup***REMOVED*** StatsChunkGroup */
/** @typedef {import("./stats/DefaultStatsFactoryPlugin").StatsChunkOrigin***REMOVED*** StatsChunkOrigin */
/** @typedef {import("./stats/DefaultStatsFactoryPlugin").StatsCompilation***REMOVED*** StatsCompilation */
/** @typedef {import("./stats/DefaultStatsFactoryPlugin").StatsError***REMOVED*** StatsError */
/** @typedef {import("./stats/DefaultStatsFactoryPlugin").StatsLogging***REMOVED*** StatsLogging */
/** @typedef {import("./stats/DefaultStatsFactoryPlugin").StatsLoggingEntry***REMOVED*** StatsLoggingEntry */
/** @typedef {import("./stats/DefaultStatsFactoryPlugin").StatsModule***REMOVED*** StatsModule */
/** @typedef {import("./stats/DefaultStatsFactoryPlugin").StatsModuleIssuer***REMOVED*** StatsModuleIssuer */
/** @typedef {import("./stats/DefaultStatsFactoryPlugin").StatsModuleReason***REMOVED*** StatsModuleReason */
/** @typedef {import("./stats/DefaultStatsFactoryPlugin").StatsModuleTraceDependency***REMOVED*** StatsModuleTraceDependency */
/** @typedef {import("./stats/DefaultStatsFactoryPlugin").StatsModuleTraceItem***REMOVED*** StatsModuleTraceItem */
/** @typedef {import("./stats/DefaultStatsFactoryPlugin").StatsProfile***REMOVED*** StatsProfile */
/** @typedef {import("./util/fs").InputFileSystem***REMOVED*** InputFileSystem */
/** @typedef {import("./util/fs").OutputFileSystem***REMOVED*** OutputFileSystem */

/**
 * @template {EXPECTED_FUNCTION***REMOVED*** T
 * @param {() => T***REMOVED*** factory factory function
 * @returns {T***REMOVED*** function
 */
const lazyFunction = factory => {
	const fac = memoize(factory);
	const f = /** @type {unknown***REMOVED*** */ (
		/**
		 * @param {...EXPECTED_ANY***REMOVED*** args args
		 * @returns {T***REMOVED*** result
		 */
		(...args) => fac()(...args)
	);
	return /** @type {T***REMOVED*** */ (f);
***REMOVED***;

/**
 * @template A
 * @template B
 * @param {A***REMOVED*** obj input a
 * @param {B***REMOVED*** exports input b
 * @returns {A & B***REMOVED*** merged
 */
const mergeExports = (obj, exports) => {
	const descriptors = Object.getOwnPropertyDescriptors(exports);
	for (const name of Object.keys(descriptors)) {
		const descriptor = descriptors[name];
		if (descriptor.get) {
			const fn = descriptor.get;
			Object.defineProperty(obj, name, {
				configurable: false,
				enumerable: true,
				get: memoize(fn)
			***REMOVED***);
		***REMOVED*** else if (typeof descriptor.value === "object") {
			Object.defineProperty(obj, name, {
				configurable: false,
				enumerable: true,
				writable: false,
				value: mergeExports({***REMOVED***, descriptor.value)
			***REMOVED***);
		***REMOVED*** else {
			throw new Error(
				"Exposed values must be either a getter or an nested object"
			);
		***REMOVED***
	***REMOVED***
	return /** @type {A & B***REMOVED*** */ (Object.freeze(obj));
***REMOVED***;

const fn = lazyFunction(() => require("./webpack"));
module.exports = mergeExports(fn, {
	get webpack() {
		return require("./webpack");
	***REMOVED***,
	/**
	 * @returns {(configuration: Configuration | Configuration[]) => void***REMOVED*** validate fn
	 */
	get validate() {
		const webpackOptionsSchemaCheck = require("../schemas/WebpackOptions.check.js");
		const getRealValidate = memoize(
			/**
			 * @returns {(configuration: Configuration | Configuration[]) => void***REMOVED*** validate fn
			 */
			() => {
				const validateSchema = require("./validateSchema");
				const webpackOptionsSchema = require("../schemas/WebpackOptions.json");
				return options => validateSchema(webpackOptionsSchema, options);
			***REMOVED***
		);
		return options => {
			if (!webpackOptionsSchemaCheck(/** @type {TODO***REMOVED*** */ (options)))
				getRealValidate()(options);
		***REMOVED***;
	***REMOVED***,
	get validateSchema() {
		const validateSchema = require("./validateSchema");
		return validateSchema;
	***REMOVED***,
	get version() {
		return /** @type {string***REMOVED*** */ (require("../package.json").version);
	***REMOVED***,

	get cli() {
		return require("./cli");
	***REMOVED***,
	get AutomaticPrefetchPlugin() {
		return require("./AutomaticPrefetchPlugin");
	***REMOVED***,
	get AsyncDependenciesBlock() {
		return require("./AsyncDependenciesBlock");
	***REMOVED***,
	get BannerPlugin() {
		return require("./BannerPlugin");
	***REMOVED***,
	get Cache() {
		return require("./Cache");
	***REMOVED***,
	get Chunk() {
		return require("./Chunk");
	***REMOVED***,
	get ChunkGraph() {
		return require("./ChunkGraph");
	***REMOVED***,
	get CleanPlugin() {
		return require("./CleanPlugin");
	***REMOVED***,
	get Compilation() {
		return require("./Compilation");
	***REMOVED***,
	get Compiler() {
		return require("./Compiler");
	***REMOVED***,
	get ConcatenationScope() {
		return require("./ConcatenationScope");
	***REMOVED***,
	get ContextExclusionPlugin() {
		return require("./ContextExclusionPlugin");
	***REMOVED***,
	get ContextReplacementPlugin() {
		return require("./ContextReplacementPlugin");
	***REMOVED***,
	get DefinePlugin() {
		return require("./DefinePlugin");
	***REMOVED***,
	get DelegatedPlugin() {
		return require("./DelegatedPlugin");
	***REMOVED***,
	get Dependency() {
		return require("./Dependency");
	***REMOVED***,
	get DllPlugin() {
		return require("./DllPlugin");
	***REMOVED***,
	get DllReferencePlugin() {
		return require("./DllReferencePlugin");
	***REMOVED***,
	get DynamicEntryPlugin() {
		return require("./DynamicEntryPlugin");
	***REMOVED***,
	get EntryOptionPlugin() {
		return require("./EntryOptionPlugin");
	***REMOVED***,
	get EntryPlugin() {
		return require("./EntryPlugin");
	***REMOVED***,
	get EnvironmentPlugin() {
		return require("./EnvironmentPlugin");
	***REMOVED***,
	get EvalDevToolModulePlugin() {
		return require("./EvalDevToolModulePlugin");
	***REMOVED***,
	get EvalSourceMapDevToolPlugin() {
		return require("./EvalSourceMapDevToolPlugin");
	***REMOVED***,
	get ExternalModule() {
		return require("./ExternalModule");
	***REMOVED***,
	get ExternalsPlugin() {
		return require("./ExternalsPlugin");
	***REMOVED***,
	get Generator() {
		return require("./Generator");
	***REMOVED***,
	get HotUpdateChunk() {
		return require("./HotUpdateChunk");
	***REMOVED***,
	get HotModuleReplacementPlugin() {
		return require("./HotModuleReplacementPlugin");
	***REMOVED***,
	get InitFragment() {
		return require("./InitFragment");
	***REMOVED***,
	get IgnorePlugin() {
		return require("./IgnorePlugin");
	***REMOVED***,
	get JavascriptModulesPlugin() {
		return util.deprecate(
			() => require("./javascript/JavascriptModulesPlugin"),
			"webpack.JavascriptModulesPlugin has moved to webpack.javascript.JavascriptModulesPlugin",
			"DEP_WEBPACK_JAVASCRIPT_MODULES_PLUGIN"
		)();
	***REMOVED***,
	get LibManifestPlugin() {
		return require("./LibManifestPlugin");
	***REMOVED***,
	get LibraryTemplatePlugin() {
		return util.deprecate(
			() => require("./LibraryTemplatePlugin"),
			"webpack.LibraryTemplatePlugin is deprecated and has been replaced by compilation.outputOptions.library or compilation.addEntry + passing a library option",
			"DEP_WEBPACK_LIBRARY_TEMPLATE_PLUGIN"
		)();
	***REMOVED***,
	get LoaderOptionsPlugin() {
		return require("./LoaderOptionsPlugin");
	***REMOVED***,
	get LoaderTargetPlugin() {
		return require("./LoaderTargetPlugin");
	***REMOVED***,
	get Module() {
		return require("./Module");
	***REMOVED***,
	get ModuleFilenameHelpers() {
		return require("./ModuleFilenameHelpers");
	***REMOVED***,
	get ModuleGraph() {
		return require("./ModuleGraph");
	***REMOVED***,
	get ModuleGraphConnection() {
		return require("./ModuleGraphConnection");
	***REMOVED***,
	get NoEmitOnErrorsPlugin() {
		return require("./NoEmitOnErrorsPlugin");
	***REMOVED***,
	get NormalModule() {
		return require("./NormalModule");
	***REMOVED***,
	get NormalModuleReplacementPlugin() {
		return require("./NormalModuleReplacementPlugin");
	***REMOVED***,
	get MultiCompiler() {
		return require("./MultiCompiler");
	***REMOVED***,
	get OptimizationStages() {
		return require("./OptimizationStages");
	***REMOVED***,
	get Parser() {
		return require("./Parser");
	***REMOVED***,
	get PlatformPlugin() {
		return require("./PlatformPlugin");
	***REMOVED***,
	get PrefetchPlugin() {
		return require("./PrefetchPlugin");
	***REMOVED***,
	get ProgressPlugin() {
		return require("./ProgressPlugin");
	***REMOVED***,
	get ProvidePlugin() {
		return require("./ProvidePlugin");
	***REMOVED***,
	get RuntimeGlobals() {
		return require("./RuntimeGlobals");
	***REMOVED***,
	get RuntimeModule() {
		return require("./RuntimeModule");
	***REMOVED***,
	get SingleEntryPlugin() {
		return util.deprecate(
			() => require("./EntryPlugin"),
			"SingleEntryPlugin was renamed to EntryPlugin",
			"DEP_WEBPACK_SINGLE_ENTRY_PLUGIN"
		)();
	***REMOVED***,
	get SourceMapDevToolPlugin() {
		return require("./SourceMapDevToolPlugin");
	***REMOVED***,
	get Stats() {
		return require("./Stats");
	***REMOVED***,
	get Template() {
		return require("./Template");
	***REMOVED***,
	get UsageState() {
		return require("./ExportsInfo").UsageState;
	***REMOVED***,
	get WatchIgnorePlugin() {
		return require("./WatchIgnorePlugin");
	***REMOVED***,
	get WebpackError() {
		return require("./WebpackError");
	***REMOVED***,
	get WebpackOptionsApply() {
		return require("./WebpackOptionsApply");
	***REMOVED***,
	get WebpackOptionsDefaulter() {
		return util.deprecate(
			() => require("./WebpackOptionsDefaulter"),
			"webpack.WebpackOptionsDefaulter is deprecated and has been replaced by webpack.config.getNormalizedWebpackOptions and webpack.config.applyWebpackOptionsDefaults",
			"DEP_WEBPACK_OPTIONS_DEFAULTER"
		)();
	***REMOVED***,
	// TODO webpack 6 deprecate
	get WebpackOptionsValidationError() {
		return require("schema-utils").ValidationError;
	***REMOVED***,
	get ValidationError() {
		return require("schema-utils").ValidationError;
	***REMOVED***,

	cache: {
		get MemoryCachePlugin() {
			return require("./cache/MemoryCachePlugin");
		***REMOVED***
	***REMOVED***,

	config: {
		get getNormalizedWebpackOptions() {
			return require("./config/normalization").getNormalizedWebpackOptions;
		***REMOVED***,
		get applyWebpackOptionsDefaults() {
			return require("./config/defaults").applyWebpackOptionsDefaults;
		***REMOVED***
	***REMOVED***,

	dependencies: {
		get ModuleDependency() {
			return require("./dependencies/ModuleDependency");
		***REMOVED***,
		get HarmonyImportDependency() {
			return require("./dependencies/HarmonyImportDependency");
		***REMOVED***,
		get ConstDependency() {
			return require("./dependencies/ConstDependency");
		***REMOVED***,
		get NullDependency() {
			return require("./dependencies/NullDependency");
		***REMOVED***
	***REMOVED***,

	ids: {
		get ChunkModuleIdRangePlugin() {
			return require("./ids/ChunkModuleIdRangePlugin");
		***REMOVED***,
		get NaturalModuleIdsPlugin() {
			return require("./ids/NaturalModuleIdsPlugin");
		***REMOVED***,
		get OccurrenceModuleIdsPlugin() {
			return require("./ids/OccurrenceModuleIdsPlugin");
		***REMOVED***,
		get NamedModuleIdsPlugin() {
			return require("./ids/NamedModuleIdsPlugin");
		***REMOVED***,
		get DeterministicChunkIdsPlugin() {
			return require("./ids/DeterministicChunkIdsPlugin");
		***REMOVED***,
		get DeterministicModuleIdsPlugin() {
			return require("./ids/DeterministicModuleIdsPlugin");
		***REMOVED***,
		get NamedChunkIdsPlugin() {
			return require("./ids/NamedChunkIdsPlugin");
		***REMOVED***,
		get OccurrenceChunkIdsPlugin() {
			return require("./ids/OccurrenceChunkIdsPlugin");
		***REMOVED***,
		get HashedModuleIdsPlugin() {
			return require("./ids/HashedModuleIdsPlugin");
		***REMOVED***
	***REMOVED***,

	javascript: {
		get EnableChunkLoadingPlugin() {
			return require("./javascript/EnableChunkLoadingPlugin");
		***REMOVED***,
		get JavascriptModulesPlugin() {
			return require("./javascript/JavascriptModulesPlugin");
		***REMOVED***,
		get JavascriptParser() {
			return require("./javascript/JavascriptParser");
		***REMOVED***
	***REMOVED***,

	optimize: {
		get AggressiveMergingPlugin() {
			return require("./optimize/AggressiveMergingPlugin");
		***REMOVED***,
		get AggressiveSplittingPlugin() {
			return util.deprecate(
				() => require("./optimize/AggressiveSplittingPlugin"),
				"AggressiveSplittingPlugin is deprecated in favor of SplitChunksPlugin",
				"DEP_WEBPACK_AGGRESSIVE_SPLITTING_PLUGIN"
			)();
		***REMOVED***,
		get InnerGraph() {
			return require("./optimize/InnerGraph");
		***REMOVED***,
		get LimitChunkCountPlugin() {
			return require("./optimize/LimitChunkCountPlugin");
		***REMOVED***,
		get MergeDuplicateChunksPlugin() {
			return require("./optimize/MergeDuplicateChunksPlugin.js");
		***REMOVED***,
		get MinChunkSizePlugin() {
			return require("./optimize/MinChunkSizePlugin");
		***REMOVED***,
		get ModuleConcatenationPlugin() {
			return require("./optimize/ModuleConcatenationPlugin");
		***REMOVED***,
		get RealContentHashPlugin() {
			return require("./optimize/RealContentHashPlugin");
		***REMOVED***,
		get RuntimeChunkPlugin() {
			return require("./optimize/RuntimeChunkPlugin");
		***REMOVED***,
		get SideEffectsFlagPlugin() {
			return require("./optimize/SideEffectsFlagPlugin");
		***REMOVED***,
		get SplitChunksPlugin() {
			return require("./optimize/SplitChunksPlugin");
		***REMOVED***
	***REMOVED***,

	runtime: {
		get GetChunkFilenameRuntimeModule() {
			return require("./runtime/GetChunkFilenameRuntimeModule");
		***REMOVED***,
		get LoadScriptRuntimeModule() {
			return require("./runtime/LoadScriptRuntimeModule");
		***REMOVED***
	***REMOVED***,

	prefetch: {
		get ChunkPrefetchPreloadPlugin() {
			return require("./prefetch/ChunkPrefetchPreloadPlugin");
		***REMOVED***
	***REMOVED***,

	web: {
		get FetchCompileWasmPlugin() {
			return require("./web/FetchCompileWasmPlugin");
		***REMOVED***,
		get FetchCompileAsyncWasmPlugin() {
			return require("./web/FetchCompileAsyncWasmPlugin");
		***REMOVED***,
		get JsonpChunkLoadingRuntimeModule() {
			return require("./web/JsonpChunkLoadingRuntimeModule");
		***REMOVED***,
		get JsonpTemplatePlugin() {
			return require("./web/JsonpTemplatePlugin");
		***REMOVED***,
		get CssLoadingRuntimeModule() {
			return require("./css/CssLoadingRuntimeModule");
		***REMOVED***
	***REMOVED***,

	esm: {
		get ModuleChunkLoadingRuntimeModule() {
			return require("./esm/ModuleChunkLoadingRuntimeModule");
		***REMOVED***
	***REMOVED***,

	webworker: {
		get WebWorkerTemplatePlugin() {
			return require("./webworker/WebWorkerTemplatePlugin");
		***REMOVED***
	***REMOVED***,

	node: {
		get NodeEnvironmentPlugin() {
			return require("./node/NodeEnvironmentPlugin");
		***REMOVED***,
		get NodeSourcePlugin() {
			return require("./node/NodeSourcePlugin");
		***REMOVED***,
		get NodeTargetPlugin() {
			return require("./node/NodeTargetPlugin");
		***REMOVED***,
		get NodeTemplatePlugin() {
			return require("./node/NodeTemplatePlugin");
		***REMOVED***,
		get ReadFileCompileWasmPlugin() {
			return require("./node/ReadFileCompileWasmPlugin");
		***REMOVED***,
		get ReadFileCompileAsyncWasmPlugin() {
			return require("./node/ReadFileCompileAsyncWasmPlugin");
		***REMOVED***
	***REMOVED***,

	electron: {
		get ElectronTargetPlugin() {
			return require("./electron/ElectronTargetPlugin");
		***REMOVED***
	***REMOVED***,

	wasm: {
		get AsyncWebAssemblyModulesPlugin() {
			return require("./wasm-async/AsyncWebAssemblyModulesPlugin");
		***REMOVED***,
		get EnableWasmLoadingPlugin() {
			return require("./wasm/EnableWasmLoadingPlugin");
		***REMOVED***
	***REMOVED***,

	css: {
		get CssModulesPlugin() {
			return require("./css/CssModulesPlugin");
		***REMOVED***
	***REMOVED***,

	library: {
		get AbstractLibraryPlugin() {
			return require("./library/AbstractLibraryPlugin");
		***REMOVED***,
		get EnableLibraryPlugin() {
			return require("./library/EnableLibraryPlugin");
		***REMOVED***
	***REMOVED***,

	container: {
		get ContainerPlugin() {
			return require("./container/ContainerPlugin");
		***REMOVED***,
		get ContainerReferencePlugin() {
			return require("./container/ContainerReferencePlugin");
		***REMOVED***,
		get ModuleFederationPlugin() {
			return require("./container/ModuleFederationPlugin");
		***REMOVED***,
		get scope() {
			return require("./container/options").scope;
		***REMOVED***
	***REMOVED***,

	sharing: {
		get ConsumeSharedPlugin() {
			return require("./sharing/ConsumeSharedPlugin");
		***REMOVED***,
		get ProvideSharedPlugin() {
			return require("./sharing/ProvideSharedPlugin");
		***REMOVED***,
		get SharePlugin() {
			return require("./sharing/SharePlugin");
		***REMOVED***,
		get scope() {
			return require("./container/options").scope;
		***REMOVED***
	***REMOVED***,

	debug: {
		get ProfilingPlugin() {
			return require("./debug/ProfilingPlugin");
		***REMOVED***
	***REMOVED***,

	util: {
		get createHash() {
			return require("./util/createHash");
		***REMOVED***,
		get comparators() {
			return require("./util/comparators");
		***REMOVED***,
		get runtime() {
			return require("./util/runtime");
		***REMOVED***,
		get serialization() {
			return require("./util/serialization");
		***REMOVED***,
		get cleverMerge() {
			return require("./util/cleverMerge").cachedCleverMerge;
		***REMOVED***,
		get LazySet() {
			return require("./util/LazySet");
		***REMOVED***,
		get compileBooleanMatcher() {
			return require("./util/compileBooleanMatcher");
		***REMOVED***
	***REMOVED***,

	get sources() {
		return require("webpack-sources");
	***REMOVED***,

	experiments: {
		schemes: {
			get HttpUriPlugin() {
				return require("./schemes/HttpUriPlugin");
			***REMOVED***
		***REMOVED***,
		ids: {
			get SyncModuleIdsPlugin() {
				return require("./ids/SyncModuleIdsPlugin");
			***REMOVED***
		***REMOVED***
	***REMOVED***
***REMOVED***);
