// webpack/lib/LoaderOptionsPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const ModuleFilenameHelpers = require("./ModuleFilenameHelpers");
const NormalModule = require("./NormalModule");
const createSchemaValidation = require("./util/create-schema-validation");

/** @typedef {import("../declarations/plugins/LoaderOptionsPlugin").LoaderOptionsPluginOptions***REMOVED*** LoaderOptionsPluginOptions */
/** @typedef {import("./Compiler")***REMOVED*** Compiler */
/** @typedef {import("./ModuleFilenameHelpers").Matcher***REMOVED*** Matcher  */
/** @typedef {import("./ModuleFilenameHelpers").MatchObject***REMOVED*** MatchObject  */

/**
 * @template T
 * @typedef {import("../declarations/LoaderContext").LoaderContext<T>***REMOVED*** LoaderContext
 */

const validate = createSchemaValidation(
	require("../schemas/plugins/LoaderOptionsPlugin.check.js"),
	() => require("../schemas/plugins/LoaderOptionsPlugin.json"),
	{
		name: "Loader Options Plugin",
		baseDataPath: "options"
	***REMOVED***
);

const PLUGIN_NAME = "LoaderOptionsPlugin";

class LoaderOptionsPlugin {
	/**
	 * @param {LoaderOptionsPluginOptions & MatchObject***REMOVED*** options options object
	 */
	constructor(options = {***REMOVED***) {
		validate(options);
		// If no options are set then generate empty options object
		if (typeof options !== "object") options = {***REMOVED***;
		if (!options.test) {
			/** @type {TODO***REMOVED*** */
			const defaultTrueMockRegExp = {
				test: () => true
			***REMOVED***;

			/** @type {RegExp***REMOVED*** */
			options.test = defaultTrueMockRegExp;
		***REMOVED***
		this.options = options;
	***REMOVED***

	/**
	 * Apply the plugin
	 * @param {Compiler***REMOVED*** compiler the compiler instance
	 * @returns {void***REMOVED***
	 */
	apply(compiler) {
		const options = this.options;
		compiler.hooks.compilation.tap(PLUGIN_NAME, compilation => {
			NormalModule.getCompilationHooks(compilation).loader.tap(
				PLUGIN_NAME,
				(context, module) => {
					const resource = module.resource;
					if (!resource) return;
					const i = resource.indexOf("?");
					if (
						ModuleFilenameHelpers.matchObject(
							options,
							i < 0 ? resource : resource.slice(0, i)
						)
					) {
						for (const key of Object.keys(options)) {
							if (key === "include" || key === "exclude" || key === "test") {
								continue;
							***REMOVED***

							/** @type {TODO***REMOVED*** */
							(context)[key] = options[key];
						***REMOVED***
					***REMOVED***
				***REMOVED***
			);
		***REMOVED***);
	***REMOVED***
***REMOVED***

module.exports = LoaderOptionsPlugin;
