// webpack/lib/ModuleGraph.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const util = require("util");
const ExportsInfo = require("./ExportsInfo");
const ModuleGraphConnection = require("./ModuleGraphConnection");
const SortableSet = require("./util/SortableSet");
const WeakTupleMap = require("./util/WeakTupleMap");

/** @typedef {import("./Compilation").ModuleMemCaches***REMOVED*** ModuleMemCaches */
/** @typedef {import("./DependenciesBlock")***REMOVED*** DependenciesBlock */
/** @typedef {import("./Dependency")***REMOVED*** Dependency */
/** @typedef {import("./ExportsInfo").ExportInfo***REMOVED*** ExportInfo */
/** @typedef {import("./Module")***REMOVED*** Module */
/** @typedef {import("./ModuleProfile")***REMOVED*** ModuleProfile */
/** @typedef {import("./RequestShortener")***REMOVED*** RequestShortener */
/** @typedef {import("./util/runtime").RuntimeSpec***REMOVED*** RuntimeSpec */

/**
 * @callback OptimizationBailoutFunction
 * @param {RequestShortener***REMOVED*** requestShortener
 * @returns {string***REMOVED***
 */

const EMPTY_SET = new Set();

/**
 * @param {SortableSet<ModuleGraphConnection>***REMOVED*** set input
 * @returns {readonly Map<Module | undefined, readonly ModuleGraphConnection[]>***REMOVED*** mapped by origin module
 */
const getConnectionsByOriginModule = set => {
	const map = new Map();
	/** @type {Module | 0***REMOVED*** */
	let lastModule = 0;
	/** @type {ModuleGraphConnection[] | undefined***REMOVED*** */
	let lastList;
	for (const connection of set) {
		const { originModule ***REMOVED*** = connection;
		if (lastModule === originModule) {
			/** @type {ModuleGraphConnection[]***REMOVED*** */
			(lastList).push(connection);
		***REMOVED*** else {
			lastModule = /** @type {Module***REMOVED*** */ (originModule);
			const list = map.get(originModule);
			if (list !== undefined) {
				lastList = list;
				list.push(connection);
			***REMOVED*** else {
				const list = [connection];
				lastList = list;
				map.set(originModule, list);
			***REMOVED***
		***REMOVED***
	***REMOVED***
	return map;
***REMOVED***;

/**
 * @param {SortableSet<ModuleGraphConnection>***REMOVED*** set input
 * @returns {readonly Map<Module | undefined, readonly ModuleGraphConnection[]>***REMOVED*** mapped by module
 */
const getConnectionsByModule = set => {
	const map = new Map();
	/** @type {Module | 0***REMOVED*** */
	let lastModule = 0;
	/** @type {ModuleGraphConnection[] | undefined***REMOVED*** */
	let lastList;
	for (const connection of set) {
		const { module ***REMOVED*** = connection;
		if (lastModule === module) {
			/** @type {ModuleGraphConnection[]***REMOVED*** */
			(lastList).push(connection);
		***REMOVED*** else {
			lastModule = module;
			const list = map.get(module);
			if (list !== undefined) {
				lastList = list;
				list.push(connection);
			***REMOVED*** else {
				const list = [connection];
				lastList = list;
				map.set(module, list);
			***REMOVED***
		***REMOVED***
	***REMOVED***
	return map;
***REMOVED***;

/** @typedef {SortableSet<ModuleGraphConnection>***REMOVED*** IncomingConnections */
/** @typedef {SortableSet<ModuleGraphConnection>***REMOVED*** OutgoingConnections */

class ModuleGraphModule {
	constructor() {
		/** @type {IncomingConnections***REMOVED*** */
		this.incomingConnections = new SortableSet();
		/** @type {OutgoingConnections | undefined***REMOVED*** */
		this.outgoingConnections = undefined;
		/** @type {Module | null | undefined***REMOVED*** */
		this.issuer = undefined;
		/** @type {(string | OptimizationBailoutFunction)[]***REMOVED*** */
		this.optimizationBailout = [];
		/** @type {ExportsInfo***REMOVED*** */
		this.exports = new ExportsInfo();
		/** @type {number | null***REMOVED*** */
		this.preOrderIndex = null;
		/** @type {number | null***REMOVED*** */
		this.postOrderIndex = null;
		/** @type {number | null***REMOVED*** */
		this.depth = null;
		/** @type {ModuleProfile | undefined***REMOVED*** */
		this.profile = undefined;
		/** @type {boolean***REMOVED*** */
		this.async = false;
		/** @type {ModuleGraphConnection[] | undefined***REMOVED*** */
		this._unassignedConnections = undefined;
	***REMOVED***
***REMOVED***

/** @typedef {(moduleGraphConnection: ModuleGraphConnection) => boolean***REMOVED*** FilterConnection */

/** @typedef {EXPECTED_OBJECT***REMOVED*** MetaKey */
/** @typedef {TODO***REMOVED*** Meta */

class ModuleGraph {
	constructor() {
		/**
		 * @type {WeakMap<Dependency, ModuleGraphConnection | null>***REMOVED***
		 * @private
		 */
		this._dependencyMap = new WeakMap();
		/**
		 * @type {Map<Module, ModuleGraphModule>***REMOVED***
		 * @private
		 */
		this._moduleMap = new Map();
		/**
		 * @type {WeakMap<MetaKey, Meta>***REMOVED***
		 * @private
		 */
		this._metaMap = new WeakMap();
		/**
		 * @type {WeakTupleMap<EXPECTED_ANY[], EXPECTED_ANY> | undefined***REMOVED***
		 * @private
		 */
		this._cache = undefined;
		/**
		 * @type {ModuleMemCaches | undefined***REMOVED***
		 * @private
		 */
		this._moduleMemCaches = undefined;

		/**
		 * @type {string | undefined***REMOVED***
		 * @private
		 */
		this._cacheStage = undefined;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @returns {ModuleGraphModule***REMOVED*** the internal module
	 */
	_getModuleGraphModule(module) {
		let mgm = this._moduleMap.get(module);
		if (mgm === undefined) {
			mgm = new ModuleGraphModule();
			this._moduleMap.set(module, mgm);
		***REMOVED***
		return mgm;
	***REMOVED***

	/**
	 * @param {Dependency***REMOVED*** dependency the dependency
	 * @param {DependenciesBlock***REMOVED*** block parent block
	 * @param {Module***REMOVED*** module parent module
	 * @param {number=***REMOVED*** indexInBlock position in block
	 * @returns {void***REMOVED***
	 */
	setParents(dependency, block, module, indexInBlock = -1) {
		dependency._parentDependenciesBlockIndex = indexInBlock;
		dependency._parentDependenciesBlock = block;
		dependency._parentModule = module;
	***REMOVED***

	/**
	 * @param {Dependency***REMOVED*** dependency the dependency
	 * @returns {Module | undefined***REMOVED*** parent module
	 */
	getParentModule(dependency) {
		return dependency._parentModule;
	***REMOVED***

	/**
	 * @param {Dependency***REMOVED*** dependency the dependency
	 * @returns {DependenciesBlock | undefined***REMOVED*** parent block
	 */
	getParentBlock(dependency) {
		return dependency._parentDependenciesBlock;
	***REMOVED***

	/**
	 * @param {Dependency***REMOVED*** dependency the dependency
	 * @returns {number***REMOVED*** index
	 */
	getParentBlockIndex(dependency) {
		return dependency._parentDependenciesBlockIndex;
	***REMOVED***

	/**
	 * @param {Module | null***REMOVED*** originModule the referencing module
	 * @param {Dependency***REMOVED*** dependency the referencing dependency
	 * @param {Module***REMOVED*** module the referenced module
	 * @returns {void***REMOVED***
	 */
	setResolvedModule(originModule, dependency, module) {
		const connection = new ModuleGraphConnection(
			originModule,
			dependency,
			module,
			undefined,
			dependency.weak,
			dependency.getCondition(this)
		);
		const connections = this._getModuleGraphModule(module).incomingConnections;
		connections.add(connection);
		if (originModule) {
			const mgm = this._getModuleGraphModule(originModule);
			if (mgm._unassignedConnections === undefined) {
				mgm._unassignedConnections = [];
			***REMOVED***
			mgm._unassignedConnections.push(connection);
			if (mgm.outgoingConnections === undefined) {
				mgm.outgoingConnections = new SortableSet();
			***REMOVED***
			mgm.outgoingConnections.add(connection);
		***REMOVED*** else {
			this._dependencyMap.set(dependency, connection);
		***REMOVED***
	***REMOVED***

	/**
	 * @param {Dependency***REMOVED*** dependency the referencing dependency
	 * @param {Module***REMOVED*** module the referenced module
	 * @returns {void***REMOVED***
	 */
	updateModule(dependency, module) {
		const connection =
			/** @type {ModuleGraphConnection***REMOVED*** */
			(this.getConnection(dependency));
		if (connection.module === module) return;
		const newConnection = connection.clone();
		newConnection.module = module;
		this._dependencyMap.set(dependency, newConnection);
		connection.setActive(false);
		const originMgm = this._getModuleGraphModule(
			/** @type {Module***REMOVED*** */ (connection.originModule)
		);
		/** @type {OutgoingConnections***REMOVED*** */
		(originMgm.outgoingConnections).add(newConnection);
		const targetMgm = this._getModuleGraphModule(module);
		targetMgm.incomingConnections.add(newConnection);
	***REMOVED***

	/**
	 * @param {Dependency***REMOVED*** dependency the referencing dependency
	 * @returns {void***REMOVED***
	 */
	removeConnection(dependency) {
		const connection =
			/** @type {ModuleGraphConnection***REMOVED*** */
			(this.getConnection(dependency));
		const targetMgm = this._getModuleGraphModule(connection.module);
		targetMgm.incomingConnections.delete(connection);
		const originMgm = this._getModuleGraphModule(
			/** @type {Module***REMOVED*** */ (connection.originModule)
		);
		/** @type {OutgoingConnections***REMOVED*** */
		(originMgm.outgoingConnections).delete(connection);
		this._dependencyMap.set(dependency, null);
	***REMOVED***

	/**
	 * @param {Dependency***REMOVED*** dependency the referencing dependency
	 * @param {string***REMOVED*** explanation an explanation
	 * @returns {void***REMOVED***
	 */
	addExplanation(dependency, explanation) {
		const connection =
			/** @type {ModuleGraphConnection***REMOVED*** */
			(this.getConnection(dependency));
		connection.addExplanation(explanation);
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** sourceModule the source module
	 * @param {Module***REMOVED*** targetModule the target module
	 * @returns {void***REMOVED***
	 */
	cloneModuleAttributes(sourceModule, targetModule) {
		const oldMgm = this._getModuleGraphModule(sourceModule);
		const newMgm = this._getModuleGraphModule(targetModule);
		newMgm.postOrderIndex = oldMgm.postOrderIndex;
		newMgm.preOrderIndex = oldMgm.preOrderIndex;
		newMgm.depth = oldMgm.depth;
		newMgm.exports = oldMgm.exports;
		newMgm.async = oldMgm.async;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @returns {void***REMOVED***
	 */
	removeModuleAttributes(module) {
		const mgm = this._getModuleGraphModule(module);
		mgm.postOrderIndex = null;
		mgm.preOrderIndex = null;
		mgm.depth = null;
		mgm.async = false;
	***REMOVED***

	/**
	 * @returns {void***REMOVED***
	 */
	removeAllModuleAttributes() {
		for (const mgm of this._moduleMap.values()) {
			mgm.postOrderIndex = null;
			mgm.preOrderIndex = null;
			mgm.depth = null;
			mgm.async = false;
		***REMOVED***
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** oldModule the old referencing module
	 * @param {Module***REMOVED*** newModule the new referencing module
	 * @param {FilterConnection***REMOVED*** filterConnection filter predicate for replacement
	 * @returns {void***REMOVED***
	 */
	moveModuleConnections(oldModule, newModule, filterConnection) {
		if (oldModule === newModule) return;
		const oldMgm = this._getModuleGraphModule(oldModule);
		const newMgm = this._getModuleGraphModule(newModule);
		// Outgoing connections
		const oldConnections = oldMgm.outgoingConnections;
		if (oldConnections !== undefined) {
			if (newMgm.outgoingConnections === undefined) {
				newMgm.outgoingConnections = new SortableSet();
			***REMOVED***
			const newConnections = newMgm.outgoingConnections;
			for (const connection of oldConnections) {
				if (filterConnection(connection)) {
					connection.originModule = newModule;
					newConnections.add(connection);
					oldConnections.delete(connection);
				***REMOVED***
			***REMOVED***
		***REMOVED***
		// Incoming connections
		const oldConnections2 = oldMgm.incomingConnections;
		const newConnections2 = newMgm.incomingConnections;
		for (const connection of oldConnections2) {
			if (filterConnection(connection)) {
				connection.module = newModule;
				newConnections2.add(connection);
				oldConnections2.delete(connection);
			***REMOVED***
		***REMOVED***
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** oldModule the old referencing module
	 * @param {Module***REMOVED*** newModule the new referencing module
	 * @param {FilterConnection***REMOVED*** filterConnection filter predicate for replacement
	 * @returns {void***REMOVED***
	 */
	copyOutgoingModuleConnections(oldModule, newModule, filterConnection) {
		if (oldModule === newModule) return;
		const oldMgm = this._getModuleGraphModule(oldModule);
		const newMgm = this._getModuleGraphModule(newModule);
		// Outgoing connections
		const oldConnections = oldMgm.outgoingConnections;
		if (oldConnections !== undefined) {
			if (newMgm.outgoingConnections === undefined) {
				newMgm.outgoingConnections = new SortableSet();
			***REMOVED***
			const newConnections = newMgm.outgoingConnections;
			for (const connection of oldConnections) {
				if (filterConnection(connection)) {
					const newConnection = connection.clone();
					newConnection.originModule = newModule;
					newConnections.add(newConnection);
					if (newConnection.module !== undefined) {
						const otherMgm = this._getModuleGraphModule(newConnection.module);
						otherMgm.incomingConnections.add(newConnection);
					***REMOVED***
				***REMOVED***
			***REMOVED***
		***REMOVED***
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the referenced module
	 * @param {string***REMOVED*** explanation an explanation why it's referenced
	 * @returns {void***REMOVED***
	 */
	addExtraReason(module, explanation) {
		const connections = this._getModuleGraphModule(module).incomingConnections;
		connections.add(new ModuleGraphConnection(null, null, module, explanation));
	***REMOVED***

	/**
	 * @param {Dependency***REMOVED*** dependency the dependency to look for a referenced module
	 * @returns {Module | null***REMOVED*** the referenced module
	 */
	getResolvedModule(dependency) {
		const connection = this.getConnection(dependency);
		return connection !== undefined ? connection.resolvedModule : null;
	***REMOVED***

	/**
	 * @param {Dependency***REMOVED*** dependency the dependency to look for a referenced module
	 * @returns {ModuleGraphConnection | undefined***REMOVED*** the connection
	 */
	getConnection(dependency) {
		const connection = this._dependencyMap.get(dependency);
		if (connection === undefined) {
			const module = this.getParentModule(dependency);
			if (module !== undefined) {
				const mgm = this._getModuleGraphModule(module);
				if (
					mgm._unassignedConnections &&
					mgm._unassignedConnections.length !== 0
				) {
					let foundConnection;
					for (const connection of mgm._unassignedConnections) {
						this._dependencyMap.set(
							/** @type {Dependency***REMOVED*** */ (connection.dependency),
							connection
						);
						if (connection.dependency === dependency)
							foundConnection = connection;
					***REMOVED***
					mgm._unassignedConnections.length = 0;
					if (foundConnection !== undefined) {
						return foundConnection;
					***REMOVED***
				***REMOVED***
			***REMOVED***
			this._dependencyMap.set(dependency, null);
			return;
		***REMOVED***
		return connection === null ? undefined : connection;
	***REMOVED***

	/**
	 * @param {Dependency***REMOVED*** dependency the dependency to look for a referenced module
	 * @returns {Module | null***REMOVED*** the referenced module
	 */
	getModule(dependency) {
		const connection = this.getConnection(dependency);
		return connection !== undefined ? connection.module : null;
	***REMOVED***

	/**
	 * @param {Dependency***REMOVED*** dependency the dependency to look for a referencing module
	 * @returns {Module | null***REMOVED*** the referencing module
	 */
	getOrigin(dependency) {
		const connection = this.getConnection(dependency);
		return connection !== undefined ? connection.originModule : null;
	***REMOVED***

	/**
	 * @param {Dependency***REMOVED*** dependency the dependency to look for a referencing module
	 * @returns {Module | null***REMOVED*** the original referencing module
	 */
	getResolvedOrigin(dependency) {
		const connection = this.getConnection(dependency);
		return connection !== undefined ? connection.resolvedOriginModule : null;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @returns {Iterable<ModuleGraphConnection>***REMOVED*** reasons why a module is included
	 */
	getIncomingConnections(module) {
		const connections = this._getModuleGraphModule(module).incomingConnections;
		return connections;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @returns {Iterable<ModuleGraphConnection>***REMOVED*** list of outgoing connections
	 */
	getOutgoingConnections(module) {
		const connections = this._getModuleGraphModule(module).outgoingConnections;
		return connections === undefined ? EMPTY_SET : connections;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @returns {readonly Map<Module | undefined | null, readonly ModuleGraphConnection[]>***REMOVED*** reasons why a module is included, in a map by source module
	 */
	getIncomingConnectionsByOriginModule(module) {
		const connections = this._getModuleGraphModule(module).incomingConnections;
		return connections.getFromUnorderedCache(getConnectionsByOriginModule);
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @returns {readonly Map<Module | undefined, readonly ModuleGraphConnection[]> | undefined***REMOVED*** connections to modules, in a map by module
	 */
	getOutgoingConnectionsByModule(module) {
		const connections = this._getModuleGraphModule(module).outgoingConnections;
		return connections === undefined
			? undefined
			: connections.getFromUnorderedCache(getConnectionsByModule);
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @returns {ModuleProfile | undefined***REMOVED*** the module profile
	 */
	getProfile(module) {
		const mgm = this._getModuleGraphModule(module);
		return mgm.profile;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @param {ModuleProfile | undefined***REMOVED*** profile the module profile
	 * @returns {void***REMOVED***
	 */
	setProfile(module, profile) {
		const mgm = this._getModuleGraphModule(module);
		mgm.profile = profile;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @returns {Module | null | undefined***REMOVED*** the issuer module
	 */
	getIssuer(module) {
		const mgm = this._getModuleGraphModule(module);
		return mgm.issuer;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @param {Module | null***REMOVED*** issuer the issuer module
	 * @returns {void***REMOVED***
	 */
	setIssuer(module, issuer) {
		const mgm = this._getModuleGraphModule(module);
		mgm.issuer = issuer;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @param {Module | null***REMOVED*** issuer the issuer module
	 * @returns {void***REMOVED***
	 */
	setIssuerIfUnset(module, issuer) {
		const mgm = this._getModuleGraphModule(module);
		if (mgm.issuer === undefined) mgm.issuer = issuer;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @returns {(string | OptimizationBailoutFunction)[]***REMOVED*** optimization bailouts
	 */
	getOptimizationBailout(module) {
		const mgm = this._getModuleGraphModule(module);
		return mgm.optimizationBailout;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @returns {true | string[] | null***REMOVED*** the provided exports
	 */
	getProvidedExports(module) {
		const mgm = this._getModuleGraphModule(module);
		return mgm.exports.getProvidedExports();
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @param {string | string[]***REMOVED*** exportName a name of an export
	 * @returns {boolean | null***REMOVED*** true, if the export is provided by the module.
	 * null, if it's unknown.
	 * false, if it's not provided.
	 */
	isExportProvided(module, exportName) {
		const mgm = this._getModuleGraphModule(module);
		const result = mgm.exports.isExportProvided(exportName);
		return result === undefined ? null : result;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @returns {ExportsInfo***REMOVED*** info about the exports
	 */
	getExportsInfo(module) {
		const mgm = this._getModuleGraphModule(module);
		return mgm.exports;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @param {string***REMOVED*** exportName the export
	 * @returns {ExportInfo***REMOVED*** info about the export
	 */
	getExportInfo(module, exportName) {
		const mgm = this._getModuleGraphModule(module);
		return mgm.exports.getExportInfo(exportName);
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @param {string***REMOVED*** exportName the export
	 * @returns {ExportInfo***REMOVED*** info about the export (do not modify)
	 */
	getReadOnlyExportInfo(module, exportName) {
		const mgm = this._getModuleGraphModule(module);
		return mgm.exports.getReadOnlyExportInfo(exportName);
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @param {RuntimeSpec***REMOVED*** runtime the runtime
	 * @returns {false | true | SortableSet<string> | null***REMOVED*** the used exports
	 * false: module is not used at all.
	 * true: the module namespace/object export is used.
	 * SortableSet<string>: these export names are used.
	 * empty SortableSet<string>: module is used but no export.
	 * null: unknown, worst case should be assumed.
	 */
	getUsedExports(module, runtime) {
		const mgm = this._getModuleGraphModule(module);
		return mgm.exports.getUsedExports(runtime);
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @returns {number | null***REMOVED*** the index of the module
	 */
	getPreOrderIndex(module) {
		const mgm = this._getModuleGraphModule(module);
		return mgm.preOrderIndex;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @returns {number | null***REMOVED*** the index of the module
	 */
	getPostOrderIndex(module) {
		const mgm = this._getModuleGraphModule(module);
		return mgm.postOrderIndex;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @param {number***REMOVED*** index the index of the module
	 * @returns {void***REMOVED***
	 */
	setPreOrderIndex(module, index) {
		const mgm = this._getModuleGraphModule(module);
		mgm.preOrderIndex = index;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @param {number***REMOVED*** index the index of the module
	 * @returns {boolean***REMOVED*** true, if the index was set
	 */
	setPreOrderIndexIfUnset(module, index) {
		const mgm = this._getModuleGraphModule(module);
		if (mgm.preOrderIndex === null) {
			mgm.preOrderIndex = index;
			return true;
		***REMOVED***
		return false;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @param {number***REMOVED*** index the index of the module
	 * @returns {void***REMOVED***
	 */
	setPostOrderIndex(module, index) {
		const mgm = this._getModuleGraphModule(module);
		mgm.postOrderIndex = index;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @param {number***REMOVED*** index the index of the module
	 * @returns {boolean***REMOVED*** true, if the index was set
	 */
	setPostOrderIndexIfUnset(module, index) {
		const mgm = this._getModuleGraphModule(module);
		if (mgm.postOrderIndex === null) {
			mgm.postOrderIndex = index;
			return true;
		***REMOVED***
		return false;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @returns {number | null***REMOVED*** the depth of the module
	 */
	getDepth(module) {
		const mgm = this._getModuleGraphModule(module);
		return mgm.depth;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @param {number***REMOVED*** depth the depth of the module
	 * @returns {void***REMOVED***
	 */
	setDepth(module, depth) {
		const mgm = this._getModuleGraphModule(module);
		mgm.depth = depth;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @param {number***REMOVED*** depth the depth of the module
	 * @returns {boolean***REMOVED*** true, if the depth was set
	 */
	setDepthIfLower(module, depth) {
		const mgm = this._getModuleGraphModule(module);
		if (mgm.depth === null || mgm.depth > depth) {
			mgm.depth = depth;
			return true;
		***REMOVED***
		return false;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @returns {boolean***REMOVED*** true, if the module is async
	 */
	isAsync(module) {
		const mgm = this._getModuleGraphModule(module);
		return mgm.async;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @returns {void***REMOVED***
	 */
	setAsync(module) {
		const mgm = this._getModuleGraphModule(module);
		mgm.async = true;
	***REMOVED***

	/**
	 * @param {MetaKey***REMOVED*** thing any thing
	 * @returns {Meta***REMOVED*** metadata
	 */
	getMeta(thing) {
		let meta = this._metaMap.get(thing);
		if (meta === undefined) {
			meta = Object.create(null);
			this._metaMap.set(thing, meta);
		***REMOVED***
		return meta;
	***REMOVED***

	/**
	 * @param {MetaKey***REMOVED*** thing any thing
	 * @returns {Meta | undefined***REMOVED*** metadata
	 */
	getMetaIfExisting(thing) {
		return this._metaMap.get(thing);
	***REMOVED***

	/**
	 * @param {string=***REMOVED*** cacheStage a persistent stage name for caching
	 */
	freeze(cacheStage) {
		this._cache = new WeakTupleMap();
		this._cacheStage = cacheStage;
	***REMOVED***

	unfreeze() {
		this._cache = undefined;
		this._cacheStage = undefined;
	***REMOVED***

	/**
	 * @template T
	 * @template R
	 * @param {(moduleGraph: ModuleGraph, ...args: T[]) => R***REMOVED*** fn computer
	 * @param {...T***REMOVED*** args arguments
	 * @returns {R***REMOVED*** computed value or cached
	 */
	cached(fn, ...args) {
		if (this._cache === undefined) return fn(this, ...args);
		return this._cache.provide(fn, ...args, () => fn(this, ...args));
	***REMOVED***

	/**
	 * @param {ModuleMemCaches***REMOVED*** moduleMemCaches mem caches for modules for better caching
	 */
	setModuleMemCaches(moduleMemCaches) {
		this._moduleMemCaches = moduleMemCaches;
	***REMOVED***

	/**
	 * @template {Dependency***REMOVED*** D
	 * @template {EXPECTED_ANY[]***REMOVED*** ARGS
	 * @template R
	 * @param {D***REMOVED*** dependency dependency
	 * @param {[...ARGS, (moduleGraph: ModuleGraph, dependency: D, ...args: ARGS) => R]***REMOVED*** args arguments, last argument is a function called with moduleGraph, dependency, ...args
	 * @returns {R***REMOVED*** computed value or cached
	 */
	dependencyCacheProvide(dependency, ...args) {
		const fn =
			/** @type {(moduleGraph: ModuleGraph, dependency: D, ...args: EXPECTED_ANY[]) => R***REMOVED*** */
			(args.pop());
		if (this._moduleMemCaches && this._cacheStage) {
			const memCache = this._moduleMemCaches.get(
				/** @type {Module***REMOVED*** */
				(this.getParentModule(dependency))
			);
			if (memCache !== undefined) {
				return memCache.provide(dependency, this._cacheStage, ...args, () =>
					fn(this, dependency, ...args)
				);
			***REMOVED***
		***REMOVED***
		if (this._cache === undefined) return fn(this, dependency, ...args);
		return this._cache.provide(dependency, ...args, () =>
			fn(this, dependency, ...args)
		);
	***REMOVED***

	// TODO remove in webpack 6
	/**
	 * @param {Module***REMOVED*** module the module
	 * @param {string***REMOVED*** deprecateMessage message for the deprecation message
	 * @param {string***REMOVED*** deprecationCode code for the deprecation
	 * @returns {ModuleGraph***REMOVED*** the module graph
	 */
	static getModuleGraphForModule(module, deprecateMessage, deprecationCode) {
		const fn = deprecateMap.get(deprecateMessage);
		if (fn) return fn(module);
		const newFn = util.deprecate(
			/**
			 * @param {Module***REMOVED*** module the module
			 * @returns {ModuleGraph***REMOVED*** the module graph
			 */
			module => {
				const moduleGraph = moduleGraphForModuleMap.get(module);
				if (!moduleGraph)
					throw new Error(
						`${
							deprecateMessage
						***REMOVED***There was no ModuleGraph assigned to the Module for backward-compat (Use the new API)`
					);
				return moduleGraph;
			***REMOVED***,
			`${deprecateMessage***REMOVED***: Use new ModuleGraph API`,
			deprecationCode
		);
		deprecateMap.set(deprecateMessage, newFn);
		return newFn(module);
	***REMOVED***

	// TODO remove in webpack 6
	/**
	 * @param {Module***REMOVED*** module the module
	 * @param {ModuleGraph***REMOVED*** moduleGraph the module graph
	 * @returns {void***REMOVED***
	 */
	static setModuleGraphForModule(module, moduleGraph) {
		moduleGraphForModuleMap.set(module, moduleGraph);
	***REMOVED***

	// TODO remove in webpack 6
	/**
	 * @param {Module***REMOVED*** module the module
	 * @returns {void***REMOVED***
	 */
	static clearModuleGraphForModule(module) {
		moduleGraphForModuleMap.delete(module);
	***REMOVED***
***REMOVED***

// TODO remove in webpack 6
/** @type {WeakMap<Module, ModuleGraph>***REMOVED*** */
const moduleGraphForModuleMap = new WeakMap();

// TODO remove in webpack 6
/** @type {Map<string, (module: Module) => ModuleGraph>***REMOVED*** */
const deprecateMap = new Map();

module.exports = ModuleGraph;
module.exports.ModuleGraphConnection = ModuleGraphConnection;
