// webpack/lib/ModuleBuildError.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const { cutOffLoaderExecution ***REMOVED*** = require("./ErrorHelpers");
const WebpackError = require("./WebpackError");
const makeSerializable = require("./util/makeSerializable");

/** @typedef {import("./serialization/ObjectMiddleware").ObjectDeserializerContext***REMOVED*** ObjectDeserializerContext */
/** @typedef {import("./serialization/ObjectMiddleware").ObjectSerializerContext***REMOVED*** ObjectSerializerContext */

/** @typedef {Error & { hideStack?: boolean ***REMOVED******REMOVED*** ErrorWithHideStack */

class ModuleBuildError extends WebpackError {
	/**
	 * @param {string | ErrorWithHideStack***REMOVED*** err error thrown
	 * @param {{from?: string|null***REMOVED******REMOVED*** info additional info
	 */
	constructor(err, { from = null ***REMOVED*** = {***REMOVED***) {
		let message = "Module build failed";
		let details;

		message += from ? ` (from ${from***REMOVED***):\n` : ": ";

		if (err !== null && typeof err === "object") {
			if (typeof err.stack === "string" && err.stack) {
				const stack = cutOffLoaderExecution(err.stack);

				if (!err.hideStack) {
					message += stack;
				***REMOVED*** else {
					details = stack;

					message +=
						typeof err.message === "string" && err.message ? err.message : err;
				***REMOVED***
			***REMOVED*** else if (typeof err.message === "string" && err.message) {
				message += err.message;
			***REMOVED*** else {
				message += String(err);
			***REMOVED***
		***REMOVED*** else {
			message += String(err);
		***REMOVED***

		super(message);

		this.name = "ModuleBuildError";
		this.details = details;
		this.error = err;
	***REMOVED***

	/**
	 * @param {ObjectSerializerContext***REMOVED*** context context
	 */
	serialize(context) {
		const { write ***REMOVED*** = context;

		write(this.error);

		super.serialize(context);
	***REMOVED***

	/**
	 * @param {ObjectDeserializerContext***REMOVED*** context context
	 */
	deserialize(context) {
		const { read ***REMOVED*** = context;

		this.error = read();

		super.deserialize(context);
	***REMOVED***
***REMOVED***

makeSerializable(ModuleBuildError, "webpack/lib/ModuleBuildError");

module.exports = ModuleBuildError;
