// webpack/lib/ModuleGraphConnection.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

/** @typedef {import("./Dependency")***REMOVED*** Dependency */
/** @typedef {import("./Dependency").GetConditionFn***REMOVED*** GetConditionFn */
/** @typedef {import("./Module")***REMOVED*** Module */
/** @typedef {import("./util/runtime").RuntimeSpec***REMOVED*** RuntimeSpec */

/**
 * Module itself is not connected, but transitive modules are connected transitively.
 */
const TRANSITIVE_ONLY = Symbol("transitive only");

/**
 * While determining the active state, this flag is used to signal a circular connection.
 */
const CIRCULAR_CONNECTION = Symbol("circular connection");

/** @typedef {boolean | typeof TRANSITIVE_ONLY | typeof CIRCULAR_CONNECTION***REMOVED*** ConnectionState */

/**
 * @param {ConnectionState***REMOVED*** a first
 * @param {ConnectionState***REMOVED*** b second
 * @returns {ConnectionState***REMOVED*** merged
 */
const addConnectionStates = (a, b) => {
	if (a === true || b === true) return true;
	if (a === false) return b;
	if (b === false) return a;
	if (a === TRANSITIVE_ONLY) return b;
	if (b === TRANSITIVE_ONLY) return a;
	return a;
***REMOVED***;

/**
 * @param {ConnectionState***REMOVED*** a first
 * @param {ConnectionState***REMOVED*** b second
 * @returns {ConnectionState***REMOVED*** intersected
 */
const intersectConnectionStates = (a, b) => {
	if (a === false || b === false) return false;
	if (a === true) return b;
	if (b === true) return a;
	if (a === CIRCULAR_CONNECTION) return b;
	if (b === CIRCULAR_CONNECTION) return a;
	return a;
***REMOVED***;

class ModuleGraphConnection {
	/**
	 * @param {Module|null***REMOVED*** originModule the referencing module
	 * @param {Dependency|null***REMOVED*** dependency the referencing dependency
	 * @param {Module***REMOVED*** module the referenced module
	 * @param {string=***REMOVED*** explanation some extra detail
	 * @param {boolean=***REMOVED*** weak the reference is weak
	 * @param {false | null | GetConditionFn | undefined***REMOVED*** condition condition for the connection
	 */
	constructor(
		originModule,
		dependency,
		module,
		explanation,
		weak = false,
		condition = undefined
	) {
		this.originModule = originModule;
		this.resolvedOriginModule = originModule;
		this.dependency = dependency;
		this.resolvedModule = module;
		this.module = module;
		this.weak = weak;
		this.conditional = Boolean(condition);
		this._active = condition !== false;
		this.condition = condition || undefined;
		/** @type {Set<string> | undefined***REMOVED*** */
		this.explanations = undefined;
		if (explanation) {
			this.explanations = new Set();
			this.explanations.add(explanation);
		***REMOVED***
	***REMOVED***

	clone() {
		const clone = new ModuleGraphConnection(
			this.resolvedOriginModule,
			this.dependency,
			this.resolvedModule,
			undefined,
			this.weak,
			this.condition
		);
		clone.originModule = this.originModule;
		clone.module = this.module;
		clone.conditional = this.conditional;
		clone._active = this._active;
		if (this.explanations) clone.explanations = new Set(this.explanations);
		return clone;
	***REMOVED***

	/**
	 * @param {GetConditionFn***REMOVED*** condition condition for the connection
	 * @returns {void***REMOVED***
	 */
	addCondition(condition) {
		if (this.conditional) {
			const old =
				/** @type {GetConditionFn***REMOVED*** */
				(this.condition);
			/** @type {GetConditionFn***REMOVED*** */
			(this.condition) = (c, r) =>
				intersectConnectionStates(old(c, r), condition(c, r));
		***REMOVED*** else if (this._active) {
			this.conditional = true;
			this.condition = condition;
		***REMOVED***
	***REMOVED***

	/**
	 * @param {string***REMOVED*** explanation the explanation to add
	 * @returns {void***REMOVED***
	 */
	addExplanation(explanation) {
		if (this.explanations === undefined) {
			this.explanations = new Set();
		***REMOVED***
		this.explanations.add(explanation);
	***REMOVED***

	get explanation() {
		if (this.explanations === undefined) return "";
		return Array.from(this.explanations).join(" ");
	***REMOVED***

	/**
	 * @param {RuntimeSpec***REMOVED*** runtime the runtime
	 * @returns {boolean***REMOVED*** true, if the connection is active
	 */
	isActive(runtime) {
		if (!this.conditional) return this._active;

		return (
			/** @type {GetConditionFn***REMOVED*** */ (this.condition)(this, runtime) !== false
		);
	***REMOVED***

	/**
	 * @param {RuntimeSpec***REMOVED*** runtime the runtime
	 * @returns {boolean***REMOVED*** true, if the connection is active
	 */
	isTargetActive(runtime) {
		if (!this.conditional) return this._active;
		return (
			/** @type {GetConditionFn***REMOVED*** */ (this.condition)(this, runtime) === true
		);
	***REMOVED***

	/**
	 * @param {RuntimeSpec***REMOVED*** runtime the runtime
	 * @returns {ConnectionState***REMOVED*** true: fully active, false: inactive, TRANSITIVE: direct module inactive, but transitive connection maybe active
	 */
	getActiveState(runtime) {
		if (!this.conditional) return this._active;
		return /** @type {GetConditionFn***REMOVED*** */ (this.condition)(this, runtime);
	***REMOVED***

	/**
	 * @param {boolean***REMOVED*** value active or not
	 * @returns {void***REMOVED***
	 */
	setActive(value) {
		this.conditional = false;
		this._active = value;
	***REMOVED***

	// TODO webpack 5 remove
	get active() {
		throw new Error("Use getActiveState instead");
	***REMOVED***

	set active(value) {
		throw new Error("Use setActive instead");
	***REMOVED***
***REMOVED***

/** @typedef {typeof TRANSITIVE_ONLY***REMOVED*** TRANSITIVE_ONLY */
/** @typedef {typeof CIRCULAR_CONNECTION***REMOVED*** CIRCULAR_CONNECTION */

module.exports = ModuleGraphConnection;
module.exports.addConnectionStates = addConnectionStates;
module.exports.TRANSITIVE_ONLY = /** @type {typeof TRANSITIVE_ONLY***REMOVED*** */ (
	TRANSITIVE_ONLY
);
module.exports.CIRCULAR_CONNECTION = /** @type {typeof CIRCULAR_CONNECTION***REMOVED*** */ (
	CIRCULAR_CONNECTION
);
