// webpack/lib/MultiCompiler.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const asyncLib = require("neo-async");
const { SyncHook, MultiHook ***REMOVED*** = require("tapable");

const ConcurrentCompilationError = require("./ConcurrentCompilationError");
const MultiStats = require("./MultiStats");
const MultiWatching = require("./MultiWatching");
const WebpackError = require("./WebpackError");
const ArrayQueue = require("./util/ArrayQueue");

/** @template T @typedef {import("tapable").AsyncSeriesHook<T>***REMOVED*** AsyncSeriesHook<T> */
/** @template T @template R @typedef {import("tapable").SyncBailHook<T, R>***REMOVED*** SyncBailHook<T, R> */
/** @typedef {import("../declarations/WebpackOptions").WatchOptions***REMOVED*** WatchOptions */
/** @typedef {import("./Compiler")***REMOVED*** Compiler */
/** @typedef {import("./Stats")***REMOVED*** Stats */
/** @typedef {import("./Watching")***REMOVED*** Watching */
/** @typedef {import("./logging/Logger").Logger***REMOVED*** Logger */
/** @typedef {import("./util/fs").InputFileSystem***REMOVED*** InputFileSystem */
/** @typedef {import("./util/fs").IntermediateFileSystem***REMOVED*** IntermediateFileSystem */
/** @typedef {import("./util/fs").OutputFileSystem***REMOVED*** OutputFileSystem */
/** @typedef {import("./util/fs").WatchFileSystem***REMOVED*** WatchFileSystem */

/**
 * @template T
 * @callback Callback
 * @param {Error | null***REMOVED*** err
 * @param {T=***REMOVED*** result
 */

/**
 * @callback RunWithDependenciesHandler
 * @param {Compiler***REMOVED*** compiler
 * @param {Callback<MultiStats>***REMOVED*** callback
 */

/**
 * @typedef {object***REMOVED*** MultiCompilerOptions
 * @property {number=***REMOVED*** parallelism how many Compilers are allows to run at the same time in parallel
 */

const CLASS_NAME = "MultiCompiler";

module.exports = class MultiCompiler {
	/**
	 * @param {Compiler[] | Record<string, Compiler>***REMOVED*** compilers child compilers
	 * @param {MultiCompilerOptions***REMOVED*** options options
	 */
	constructor(compilers, options) {
		if (!Array.isArray(compilers)) {
			/** @type {Compiler[]***REMOVED*** */
			compilers = Object.keys(compilers).map(name => {
				/** @type {Record<string, Compiler>***REMOVED*** */
				(compilers)[name].name = name;
				return /** @type {Record<string, Compiler>***REMOVED*** */ (compilers)[name];
			***REMOVED***);
		***REMOVED***

		this.hooks = Object.freeze({
			/** @type {SyncHook<[MultiStats]>***REMOVED*** */
			done: new SyncHook(["stats"]),
			/** @type {MultiHook<SyncHook<[string | null, number]>>***REMOVED*** */
			invalid: new MultiHook(compilers.map(c => c.hooks.invalid)),
			/** @type {MultiHook<AsyncSeriesHook<[Compiler]>>***REMOVED*** */
			run: new MultiHook(compilers.map(c => c.hooks.run)),
			/** @type {SyncHook<[]>***REMOVED*** */
			watchClose: new SyncHook([]),
			/** @type {MultiHook<AsyncSeriesHook<[Compiler]>>***REMOVED*** */
			watchRun: new MultiHook(compilers.map(c => c.hooks.watchRun)),
			/** @type {MultiHook<SyncBailHook<[string, string, EXPECTED_ANY[] | undefined], true | void>>***REMOVED*** */
			infrastructureLog: new MultiHook(
				compilers.map(c => c.hooks.infrastructureLog)
			)
		***REMOVED***);
		this.compilers = compilers;
		/** @type {MultiCompilerOptions***REMOVED*** */
		this._options = {
			parallelism: options.parallelism || Infinity
		***REMOVED***;
		/** @type {WeakMap<Compiler, string[]>***REMOVED*** */
		this.dependencies = new WeakMap();
		this.running = false;

		/** @type {(Stats | null)[]***REMOVED*** */
		const compilerStats = this.compilers.map(() => null);
		let doneCompilers = 0;
		for (let index = 0; index < this.compilers.length; index++) {
			const compiler = this.compilers[index];
			const compilerIndex = index;
			let compilerDone = false;
			// eslint-disable-next-line no-loop-func
			compiler.hooks.done.tap(CLASS_NAME, stats => {
				if (!compilerDone) {
					compilerDone = true;
					doneCompilers++;
				***REMOVED***
				compilerStats[compilerIndex] = stats;
				if (doneCompilers === this.compilers.length) {
					this.hooks.done.call(
						new MultiStats(/** @type {Stats[]***REMOVED*** */ (compilerStats))
					);
				***REMOVED***
			***REMOVED***);
			// eslint-disable-next-line no-loop-func
			compiler.hooks.invalid.tap(CLASS_NAME, () => {
				if (compilerDone) {
					compilerDone = false;
					doneCompilers--;
				***REMOVED***
			***REMOVED***);
		***REMOVED***
		this._validateCompilersOptions();
	***REMOVED***

	_validateCompilersOptions() {
		if (this.compilers.length < 2) return;
		/**
		 * @param {Compiler***REMOVED*** compiler compiler
		 * @param {WebpackError***REMOVED*** warning warning
		 */
		const addWarning = (compiler, warning) => {
			compiler.hooks.thisCompilation.tap(CLASS_NAME, compilation => {
				compilation.warnings.push(warning);
			***REMOVED***);
		***REMOVED***;
		const cacheNames = new Set();
		for (const compiler of this.compilers) {
			if (compiler.options.cache && "name" in compiler.options.cache) {
				const name = compiler.options.cache.name;
				if (cacheNames.has(name)) {
					addWarning(
						compiler,
						new WebpackError(
							`${
								compiler.name
									? `Compiler with name "${compiler.name***REMOVED***" doesn't use unique cache name. `
									: ""
							***REMOVED***Please set unique "cache.name" option. Name "${name***REMOVED***" already used.`
						)
					);
				***REMOVED*** else {
					cacheNames.add(name);
				***REMOVED***
			***REMOVED***
		***REMOVED***
	***REMOVED***

	get options() {
		return Object.assign(
			this.compilers.map(c => c.options),
			this._options
		);
	***REMOVED***

	get outputPath() {
		let commonPath = this.compilers[0].outputPath;
		for (const compiler of this.compilers) {
			while (
				compiler.outputPath.indexOf(commonPath) !== 0 &&
				/[/\\]/.test(commonPath)
			) {
				commonPath = commonPath.replace(/[/\\][^/\\]*$/, "");
			***REMOVED***
		***REMOVED***

		if (!commonPath && this.compilers[0].outputPath[0] === "/") return "/";
		return commonPath;
	***REMOVED***

	get inputFileSystem() {
		throw new Error("Cannot read inputFileSystem of a MultiCompiler");
	***REMOVED***

	/**
	 * @param {InputFileSystem***REMOVED*** value the new input file system
	 */
	set inputFileSystem(value) {
		for (const compiler of this.compilers) {
			compiler.inputFileSystem = value;
		***REMOVED***
	***REMOVED***

	get outputFileSystem() {
		throw new Error("Cannot read outputFileSystem of a MultiCompiler");
	***REMOVED***

	/**
	 * @param {OutputFileSystem***REMOVED*** value the new output file system
	 */
	set outputFileSystem(value) {
		for (const compiler of this.compilers) {
			compiler.outputFileSystem = value;
		***REMOVED***
	***REMOVED***

	get watchFileSystem() {
		throw new Error("Cannot read watchFileSystem of a MultiCompiler");
	***REMOVED***

	/**
	 * @param {WatchFileSystem***REMOVED*** value the new watch file system
	 */
	set watchFileSystem(value) {
		for (const compiler of this.compilers) {
			compiler.watchFileSystem = value;
		***REMOVED***
	***REMOVED***

	/**
	 * @param {IntermediateFileSystem***REMOVED*** value the new intermediate file system
	 */
	set intermediateFileSystem(value) {
		for (const compiler of this.compilers) {
			compiler.intermediateFileSystem = value;
		***REMOVED***
	***REMOVED***

	get intermediateFileSystem() {
		throw new Error("Cannot read outputFileSystem of a MultiCompiler");
	***REMOVED***

	/**
	 * @param {string | (() => string)***REMOVED*** name name of the logger, or function called once to get the logger name
	 * @returns {Logger***REMOVED*** a logger with that name
	 */
	getInfrastructureLogger(name) {
		return this.compilers[0].getInfrastructureLogger(name);
	***REMOVED***

	/**
	 * @param {Compiler***REMOVED*** compiler the child compiler
	 * @param {string[]***REMOVED*** dependencies its dependencies
	 * @returns {void***REMOVED***
	 */
	setDependencies(compiler, dependencies) {
		this.dependencies.set(compiler, dependencies);
	***REMOVED***

	/**
	 * @param {Callback<MultiStats>***REMOVED*** callback signals when the validation is complete
	 * @returns {boolean***REMOVED*** true if the dependencies are valid
	 */
	validateDependencies(callback) {
		/** @type {Set<{source: Compiler, target: Compiler***REMOVED***>***REMOVED*** */
		const edges = new Set();
		/** @type {string[]***REMOVED*** */
		const missing = [];
		/**
		 * @param {Compiler***REMOVED*** compiler compiler
		 * @returns {boolean***REMOVED*** target was found
		 */
		const targetFound = compiler => {
			for (const edge of edges) {
				if (edge.target === compiler) {
					return true;
				***REMOVED***
			***REMOVED***
			return false;
		***REMOVED***;
		/**
		 * @param {{source: Compiler, target: Compiler***REMOVED******REMOVED*** e1 edge 1
		 * @param {{source: Compiler, target: Compiler***REMOVED******REMOVED*** e2 edge 2
		 * @returns {number***REMOVED*** result
		 */
		const sortEdges = (e1, e2) =>
			/** @type {string***REMOVED*** */
			(e1.source.name).localeCompare(/** @type {string***REMOVED*** */ (e2.source.name)) ||
			/** @type {string***REMOVED*** */
			(e1.target.name).localeCompare(/** @type {string***REMOVED*** */ (e2.target.name));
		for (const source of this.compilers) {
			const dependencies = this.dependencies.get(source);
			if (dependencies) {
				for (const dep of dependencies) {
					const target = this.compilers.find(c => c.name === dep);
					if (!target) {
						missing.push(dep);
					***REMOVED*** else {
						edges.add({
							source,
							target
						***REMOVED***);
					***REMOVED***
				***REMOVED***
			***REMOVED***
		***REMOVED***
		/** @type {string[]***REMOVED*** */
		const errors = missing.map(m => `Compiler dependency \`${m***REMOVED***\` not found.`);
		const stack = this.compilers.filter(c => !targetFound(c));
		while (stack.length > 0) {
			const current = stack.pop();
			for (const edge of edges) {
				if (edge.source === current) {
					edges.delete(edge);
					const target = edge.target;
					if (!targetFound(target)) {
						stack.push(target);
					***REMOVED***
				***REMOVED***
			***REMOVED***
		***REMOVED***
		if (edges.size > 0) {
			/** @type {string[]***REMOVED*** */
			const lines = Array.from(edges)
				.sort(sortEdges)
				.map(edge => `${edge.source.name***REMOVED*** -> ${edge.target.name***REMOVED***`);
			lines.unshift("Circular dependency found in compiler dependencies.");
			errors.unshift(lines.join("\n"));
		***REMOVED***
		if (errors.length > 0) {
			const message = errors.join("\n");
			callback(new Error(message));
			return false;
		***REMOVED***
		return true;
	***REMOVED***

	// TODO webpack 6 remove
	/**
	 * @deprecated This method should have been private
	 * @param {Compiler[]***REMOVED*** compilers the child compilers
	 * @param {RunWithDependenciesHandler***REMOVED*** fn a handler to run for each compiler
	 * @param {Callback<MultiStats>***REMOVED*** callback the compiler's handler
	 * @returns {void***REMOVED***
	 */
	runWithDependencies(compilers, fn, callback) {
		const fulfilledNames = new Set();
		let remainingCompilers = compilers;
		/**
		 * @param {string***REMOVED*** d dependency
		 * @returns {boolean***REMOVED*** when dependency was fulfilled
		 */
		const isDependencyFulfilled = d => fulfilledNames.has(d);
		/**
		 * @returns {Compiler[]***REMOVED*** compilers
		 */
		const getReadyCompilers = () => {
			const readyCompilers = [];
			const list = remainingCompilers;
			remainingCompilers = [];
			for (const c of list) {
				const dependencies = this.dependencies.get(c);
				const ready =
					!dependencies || dependencies.every(isDependencyFulfilled);
				if (ready) {
					readyCompilers.push(c);
				***REMOVED*** else {
					remainingCompilers.push(c);
				***REMOVED***
			***REMOVED***
			return readyCompilers;
		***REMOVED***;
		/**
		 * @param {Callback<MultiStats>***REMOVED*** callback callback
		 * @returns {void***REMOVED***
		 */
		const runCompilers = callback => {
			if (remainingCompilers.length === 0) return callback(null);
			asyncLib.map(
				getReadyCompilers(),
				(compiler, callback) => {
					fn(compiler, err => {
						if (err) return callback(err);
						fulfilledNames.add(compiler.name);
						runCompilers(callback);
					***REMOVED***);
				***REMOVED***,
				(err, results) => {
					callback(err, /** @type {TODO***REMOVED*** */ (results));
				***REMOVED***
			);
		***REMOVED***;
		runCompilers(callback);
	***REMOVED***

	/**
	 * @template SetupResult
	 * @param {(compiler: Compiler, index: number, doneCallback: Callback<Stats>, isBlocked: () => boolean, setChanged: () => void, setInvalid: () => void) => SetupResult***REMOVED*** setup setup a single compiler
	 * @param {(compiler: Compiler, setupResult: SetupResult, callback: Callback<Stats>) => void***REMOVED*** run run/continue a single compiler
	 * @param {Callback<MultiStats>***REMOVED*** callback callback when all compilers are done, result includes Stats of all changed compilers
	 * @returns {SetupResult[]***REMOVED*** result of setup
	 */
	_runGraph(setup, run, callback) {
		/** @typedef {{ compiler: Compiler, setupResult: undefined | SetupResult, result: undefined | Stats, state: "pending" | "blocked" | "queued" | "starting" | "running" | "running-outdated" | "done", children: Node[], parents: Node[] ***REMOVED******REMOVED*** Node */

		// State transitions for nodes:
		// -> blocked (initial)
		// blocked -> starting [running++] (when all parents done)
		// queued -> starting [running++] (when processing the queue)
		// starting -> running (when run has been called)
		// running -> done [running--] (when compilation is done)
		// done -> pending (when invalidated from file change)
		// pending -> blocked [add to queue] (when invalidated from aggregated changes)
		// done -> blocked [add to queue] (when invalidated, from parent invalidation)
		// running -> running-outdated (when invalidated, either from change or parent invalidation)
		// running-outdated -> blocked [running--] (when compilation is done)

		/** @type {Node[]***REMOVED*** */
		const nodes = this.compilers.map(compiler => ({
			compiler,
			setupResult: undefined,
			result: undefined,
			state: "blocked",
			children: [],
			parents: []
		***REMOVED***));
		/** @type {Map<string, Node>***REMOVED*** */
		const compilerToNode = new Map();
		for (const node of nodes) {
			compilerToNode.set(/** @type {string***REMOVED*** */ (node.compiler.name), node);
		***REMOVED***
		for (const node of nodes) {
			const dependencies = this.dependencies.get(node.compiler);
			if (!dependencies) continue;
			for (const dep of dependencies) {
				const parent = /** @type {Node***REMOVED*** */ (compilerToNode.get(dep));
				node.parents.push(parent);
				parent.children.push(node);
			***REMOVED***
		***REMOVED***
		/** @type {ArrayQueue<Node>***REMOVED*** */
		const queue = new ArrayQueue();
		for (const node of nodes) {
			if (node.parents.length === 0) {
				node.state = "queued";
				queue.enqueue(node);
			***REMOVED***
		***REMOVED***
		let errored = false;
		let running = 0;
		const parallelism = /** @type {number***REMOVED*** */ (this._options.parallelism);
		/**
		 * @param {Node***REMOVED*** node node
		 * @param {(Error | null)=***REMOVED*** err error
		 * @param {Stats=***REMOVED*** stats result
		 * @returns {void***REMOVED***
		 */
		const nodeDone = (node, err, stats) => {
			if (errored) return;
			if (err) {
				errored = true;
				return asyncLib.each(
					nodes,
					(node, callback) => {
						if (node.compiler.watching) {
							node.compiler.watching.close(callback);
						***REMOVED*** else {
							callback();
						***REMOVED***
					***REMOVED***,
					() => callback(err)
				);
			***REMOVED***
			node.result = stats;
			running--;
			if (node.state === "running") {
				node.state = "done";
				for (const child of node.children) {
					if (child.state === "blocked") queue.enqueue(child);
				***REMOVED***
			***REMOVED*** else if (node.state === "running-outdated") {
				node.state = "blocked";
				queue.enqueue(node);
			***REMOVED***
			processQueue();
		***REMOVED***;
		/**
		 * @param {Node***REMOVED*** node node
		 * @returns {void***REMOVED***
		 */
		const nodeInvalidFromParent = node => {
			if (node.state === "done") {
				node.state = "blocked";
			***REMOVED*** else if (node.state === "running") {
				node.state = "running-outdated";
			***REMOVED***
			for (const child of node.children) {
				nodeInvalidFromParent(child);
			***REMOVED***
		***REMOVED***;
		/**
		 * @param {Node***REMOVED*** node node
		 * @returns {void***REMOVED***
		 */
		const nodeInvalid = node => {
			if (node.state === "done") {
				node.state = "pending";
			***REMOVED*** else if (node.state === "running") {
				node.state = "running-outdated";
			***REMOVED***
			for (const child of node.children) {
				nodeInvalidFromParent(child);
			***REMOVED***
		***REMOVED***;
		/**
		 * @param {Node***REMOVED*** node node
		 * @returns {void***REMOVED***
		 */
		const nodeChange = node => {
			nodeInvalid(node);
			if (node.state === "pending") {
				node.state = "blocked";
			***REMOVED***
			if (node.state === "blocked") {
				queue.enqueue(node);
				processQueue();
			***REMOVED***
		***REMOVED***;

		/** @type {SetupResult[]***REMOVED*** */
		const setupResults = [];
		for (const [i, node] of nodes.entries()) {
			setupResults.push(
				(node.setupResult = setup(
					node.compiler,
					i,
					nodeDone.bind(null, node),
					() => node.state !== "starting" && node.state !== "running",
					() => nodeChange(node),
					() => nodeInvalid(node)
				))
			);
		***REMOVED***
		let processing = true;
		const processQueue = () => {
			if (processing) return;
			processing = true;
			process.nextTick(processQueueWorker);
		***REMOVED***;
		const processQueueWorker = () => {
			// eslint-disable-next-line no-unmodified-loop-condition
			while (running < parallelism && queue.length > 0 && !errored) {
				const node = /** @type {Node***REMOVED*** */ (queue.dequeue());
				if (
					node.state === "queued" ||
					(node.state === "blocked" &&
						node.parents.every(p => p.state === "done"))
				) {
					running++;
					node.state = "starting";
					run(
						node.compiler,
						/** @type {SetupResult***REMOVED*** */ (node.setupResult),
						nodeDone.bind(null, node)
					);
					node.state = "running";
				***REMOVED***
			***REMOVED***
			processing = false;
			if (
				!errored &&
				running === 0 &&
				nodes.every(node => node.state === "done")
			) {
				const stats = [];
				for (const node of nodes) {
					const result = node.result;
					if (result) {
						node.result = undefined;
						stats.push(result);
					***REMOVED***
				***REMOVED***
				if (stats.length > 0) {
					callback(null, new MultiStats(stats));
				***REMOVED***
			***REMOVED***
		***REMOVED***;
		processQueueWorker();
		return setupResults;
	***REMOVED***

	/**
	 * @param {WatchOptions|WatchOptions[]***REMOVED*** watchOptions the watcher's options
	 * @param {Callback<MultiStats>***REMOVED*** handler signals when the call finishes
	 * @returns {MultiWatching***REMOVED*** a compiler watcher
	 */
	watch(watchOptions, handler) {
		if (this.running) {
			return handler(new ConcurrentCompilationError());
		***REMOVED***
		this.running = true;

		if (this.validateDependencies(handler)) {
			const watchings = this._runGraph(
				(compiler, idx, callback, isBlocked, setChanged, setInvalid) => {
					const watching = compiler.watch(
						Array.isArray(watchOptions) ? watchOptions[idx] : watchOptions,
						callback
					);
					if (watching) {
						watching._onInvalid = setInvalid;
						watching._onChange = setChanged;
						watching._isBlocked = isBlocked;
					***REMOVED***
					return watching;
				***REMOVED***,
				(compiler, watching, callback) => {
					if (compiler.watching !== watching) return;
					if (!watching.running) watching.invalidate();
				***REMOVED***,
				handler
			);
			return new MultiWatching(watchings, this);
		***REMOVED***

		return new MultiWatching([], this);
	***REMOVED***

	/**
	 * @param {Callback<MultiStats>***REMOVED*** callback signals when the call finishes
	 * @returns {void***REMOVED***
	 */
	run(callback) {
		if (this.running) {
			return callback(new ConcurrentCompilationError());
		***REMOVED***
		this.running = true;

		if (this.validateDependencies(callback)) {
			this._runGraph(
				() => {***REMOVED***,
				(compiler, setupResult, callback) => compiler.run(callback),
				(err, stats) => {
					this.running = false;

					if (callback !== undefined) {
						return callback(err, stats);
					***REMOVED***
				***REMOVED***
			);
		***REMOVED***
	***REMOVED***

	purgeInputFileSystem() {
		for (const compiler of this.compilers) {
			if (compiler.inputFileSystem && compiler.inputFileSystem.purge) {
				compiler.inputFileSystem.purge();
			***REMOVED***
		***REMOVED***
	***REMOVED***

	/**
	 * @param {Callback<void>***REMOVED*** callback signals when the compiler closes
	 * @returns {void***REMOVED***
	 */
	close(callback) {
		asyncLib.each(
			this.compilers,
			(compiler, callback) => {
				compiler.close(callback);
			***REMOVED***,
			error => {
				callback(error);
			***REMOVED***
		);
	***REMOVED***
***REMOVED***;
