// webpack/lib/NormalModuleFactory.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const { getContext ***REMOVED*** = require("loader-runner");
const asyncLib = require("neo-async");
const {
	AsyncSeriesBailHook,
	SyncWaterfallHook,
	SyncBailHook,
	SyncHook,
	HookMap
***REMOVED*** = require("tapable");
const ChunkGraph = require("./ChunkGraph");
const Module = require("./Module");
const ModuleFactory = require("./ModuleFactory");
const ModuleGraph = require("./ModuleGraph");
const { JAVASCRIPT_MODULE_TYPE_AUTO ***REMOVED*** = require("./ModuleTypeConstants");
const NormalModule = require("./NormalModule");
const BasicEffectRulePlugin = require("./rules/BasicEffectRulePlugin");
const BasicMatcherRulePlugin = require("./rules/BasicMatcherRulePlugin");
const ObjectMatcherRulePlugin = require("./rules/ObjectMatcherRulePlugin");
const RuleSetCompiler = require("./rules/RuleSetCompiler");
const UseEffectRulePlugin = require("./rules/UseEffectRulePlugin");
const LazySet = require("./util/LazySet");
const { getScheme ***REMOVED*** = require("./util/URLAbsoluteSpecifier");
const { cachedCleverMerge, cachedSetProperty ***REMOVED*** = require("./util/cleverMerge");
const { join ***REMOVED*** = require("./util/fs");
const {
	parseResource,
	parseResourceWithoutFragment
***REMOVED*** = require("./util/identifier");

/** @typedef {import("../declarations/WebpackOptions").ModuleOptionsNormalized***REMOVED*** ModuleOptions */
/** @typedef {import("../declarations/WebpackOptions").RuleSetRule***REMOVED*** RuleSetRule */
/** @typedef {import("./Generator")***REMOVED*** Generator */
/** @typedef {import("./ModuleFactory").ModuleFactoryCallback***REMOVED*** ModuleFactoryCallback */
/** @typedef {import("./ModuleFactory").ModuleFactoryCreateData***REMOVED*** ModuleFactoryCreateData */
/** @typedef {import("./ModuleFactory").ModuleFactoryCreateDataContextInfo***REMOVED*** ModuleFactoryCreateDataContextInfo */
/** @typedef {import("./ModuleFactory").ModuleFactoryResult***REMOVED*** ModuleFactoryResult */
/** @typedef {import("./NormalModule").GeneratorOptions***REMOVED*** GeneratorOptions */
/** @typedef {import("./NormalModule").LoaderItem***REMOVED*** LoaderItem */
/** @typedef {import("./NormalModule").NormalModuleCreateData***REMOVED*** NormalModuleCreateData */
/** @typedef {import("./NormalModule").ParserOptions***REMOVED*** ParserOptions */
/** @typedef {import("./Parser")***REMOVED*** Parser */
/** @typedef {import("./ResolverFactory")***REMOVED*** ResolverFactory */
/** @typedef {import("./ResolverFactory").ResolveContext***REMOVED*** ResolveContext */
/** @typedef {import("./ResolverFactory").ResolveRequest***REMOVED*** ResolveRequest */
/** @typedef {import("./ResolverFactory").ResolverWithOptions***REMOVED*** ResolverWithOptions */
/** @typedef {import("./dependencies/ModuleDependency")***REMOVED*** ModuleDependency */
/** @typedef {import("./javascript/JavascriptParser").ImportAttributes***REMOVED*** ImportAttributes */
/** @typedef {import("./rules/RuleSetCompiler").RuleSetRules***REMOVED*** RuleSetRules */
/** @typedef {import("./util/fs").InputFileSystem***REMOVED*** InputFileSystem */
/** @typedef {import("./util/identifier").AssociatedObjectForCache***REMOVED*** AssociatedObjectForCache */

/** @typedef {Pick<RuleSetRule, 'type' | 'sideEffects' | 'parser' | 'generator' | 'resolve' | 'layer'>***REMOVED*** ModuleSettings */
/** @typedef {Partial<NormalModuleCreateData & { settings: ModuleSettings ***REMOVED***>***REMOVED*** CreateData */

/**
 * @typedef {object***REMOVED*** ResolveData
 * @property {ModuleFactoryCreateData["contextInfo"]***REMOVED*** contextInfo
 * @property {ModuleFactoryCreateData["resolveOptions"]***REMOVED*** resolveOptions
 * @property {string***REMOVED*** context
 * @property {string***REMOVED*** request
 * @property {ImportAttributes | undefined***REMOVED*** assertions
 * @property {ModuleDependency[]***REMOVED*** dependencies
 * @property {string***REMOVED*** dependencyType
 * @property {CreateData***REMOVED*** createData
 * @property {LazySet<string>***REMOVED*** fileDependencies
 * @property {LazySet<string>***REMOVED*** missingDependencies
 * @property {LazySet<string>***REMOVED*** contextDependencies
 * @property {Module=***REMOVED*** ignoredModule
 * @property {boolean***REMOVED*** cacheable allow to use the unsafe cache
 */

/**
 * @typedef {object***REMOVED*** ResourceData
 * @property {string***REMOVED*** resource
 * @property {string=***REMOVED*** path
 * @property {string=***REMOVED*** query
 * @property {string=***REMOVED*** fragment
 * @property {string=***REMOVED*** context
 */

/** @typedef {ResourceData & { data: Record<string, EXPECTED_ANY> ***REMOVED******REMOVED*** ResourceDataWithData */

/**
 * @typedef {object***REMOVED*** ParsedLoaderRequest
 * @property {string***REMOVED*** loader loader
 * @property {string|undefined***REMOVED*** options options
 */

/**
 * @template T
 * @callback Callback
 * @param {(Error | null)=***REMOVED*** err
 * @param {T=***REMOVED*** stats
 * @returns {void***REMOVED***
 */

const EMPTY_RESOLVE_OPTIONS = {***REMOVED***;
/** @type {ParserOptions***REMOVED*** */
const EMPTY_PARSER_OPTIONS = {***REMOVED***;
/** @type {GeneratorOptions***REMOVED*** */
const EMPTY_GENERATOR_OPTIONS = {***REMOVED***;
/** @type {ParsedLoaderRequest[]***REMOVED*** */
const EMPTY_ELEMENTS = [];

const MATCH_RESOURCE_REGEX = /^([^!]+)!=!/;
const LEADING_DOT_EXTENSION_REGEX = /^[^.]/;

/**
 * @param {LoaderItem***REMOVED*** data data
 * @returns {string***REMOVED*** ident
 */
const loaderToIdent = data => {
	if (!data.options) {
		return data.loader;
	***REMOVED***
	if (typeof data.options === "string") {
		return `${data.loader***REMOVED***?${data.options***REMOVED***`;
	***REMOVED***
	if (typeof data.options !== "object") {
		throw new Error("loader options must be string or object");
	***REMOVED***
	if (data.ident) {
		return `${data.loader***REMOVED***??${data.ident***REMOVED***`;
	***REMOVED***
	return `${data.loader***REMOVED***?${JSON.stringify(data.options)***REMOVED***`;
***REMOVED***;

/**
 * @param {LoaderItem[]***REMOVED*** loaders loaders
 * @param {string***REMOVED*** resource resource
 * @returns {string***REMOVED*** stringified loaders and resource
 */
const stringifyLoadersAndResource = (loaders, resource) => {
	let str = "";
	for (const loader of loaders) {
		str += `${loaderToIdent(loader)***REMOVED***!`;
	***REMOVED***
	return str + resource;
***REMOVED***;

/**
 * @param {number***REMOVED*** times times
 * @param {(err?: null | Error) => void***REMOVED*** callback callback
 * @returns {(err?: null | Error) => void***REMOVED*** callback
 */
const needCalls = (times, callback) => err => {
	if (--times === 0) {
		return callback(err);
	***REMOVED***
	if (err && times > 0) {
		times = Number.NaN;
		return callback(err);
	***REMOVED***
***REMOVED***;

/**
 * @template T
 * @template O
 * @param {T***REMOVED*** globalOptions global options
 * @param {string***REMOVED*** type type
 * @param {O***REMOVED*** localOptions local options
 * @returns {T & O | T | O***REMOVED*** result
 */
const mergeGlobalOptions = (globalOptions, type, localOptions) => {
	const parts = type.split("/");
	let result;
	let current = "";
	for (const part of parts) {
		current = current ? `${current***REMOVED***/${part***REMOVED***` : part;
		const options =
			/** @type {T***REMOVED*** */
			(globalOptions[/** @type {keyof T***REMOVED*** */ (current)]);
		if (typeof options === "object") {
			result =
				result === undefined ? options : cachedCleverMerge(result, options);
		***REMOVED***
	***REMOVED***
	if (result === undefined) {
		return localOptions;
	***REMOVED***
	return cachedCleverMerge(result, localOptions);
***REMOVED***;

// TODO webpack 6 remove
/**
 * @template {import("tapable").Hook<EXPECTED_ANY, EXPECTED_ANY>***REMOVED*** T
 * @param {string***REMOVED*** name name
 * @param {T***REMOVED*** hook hook
 * @returns {string***REMOVED*** result
 */
const deprecationChangedHookMessage = (name, hook) => {
	const names = hook.taps.map(tapped => tapped.name).join(", ");

	return (
		`NormalModuleFactory.${name***REMOVED*** (${names***REMOVED***) is no longer a waterfall hook, but a bailing hook instead. ` +
		"Do not return the passed object, but modify it instead. " +
		"Returning false will ignore the request and results in no module created."
	);
***REMOVED***;

const ruleSetCompiler = new RuleSetCompiler([
	new BasicMatcherRulePlugin("test", "resource"),
	new BasicMatcherRulePlugin("scheme"),
	new BasicMatcherRulePlugin("mimetype"),
	new BasicMatcherRulePlugin("dependency"),
	new BasicMatcherRulePlugin("include", "resource"),
	new BasicMatcherRulePlugin("exclude", "resource", true),
	new BasicMatcherRulePlugin("resource"),
	new BasicMatcherRulePlugin("resourceQuery"),
	new BasicMatcherRulePlugin("resourceFragment"),
	new BasicMatcherRulePlugin("realResource"),
	new BasicMatcherRulePlugin("issuer"),
	new BasicMatcherRulePlugin("compiler"),
	new BasicMatcherRulePlugin("issuerLayer"),
	new ObjectMatcherRulePlugin("assert", "assertions", value => {
		if (value) {
			return (
				/** @type {ImportAttributes***REMOVED*** */ (value)._isLegacyAssert !== undefined
			);
		***REMOVED***

		return false;
	***REMOVED***),
	new ObjectMatcherRulePlugin("with", "assertions", value => {
		if (value) {
			return !(/** @type {ImportAttributes***REMOVED*** */ (value)._isLegacyAssert);
		***REMOVED***
		return false;
	***REMOVED***),
	new ObjectMatcherRulePlugin("descriptionData"),
	new BasicEffectRulePlugin("type"),
	new BasicEffectRulePlugin("sideEffects"),
	new BasicEffectRulePlugin("parser"),
	new BasicEffectRulePlugin("resolve"),
	new BasicEffectRulePlugin("generator"),
	new BasicEffectRulePlugin("layer"),
	new UseEffectRulePlugin()
]);

class NormalModuleFactory extends ModuleFactory {
	/**
	 * @param {object***REMOVED*** param params
	 * @param {string=***REMOVED*** param.context context
	 * @param {InputFileSystem***REMOVED*** param.fs file system
	 * @param {ResolverFactory***REMOVED*** param.resolverFactory resolverFactory
	 * @param {ModuleOptions***REMOVED*** param.options options
	 * @param {AssociatedObjectForCache***REMOVED*** param.associatedObjectForCache an object to which the cache will be attached
	 * @param {boolean=***REMOVED*** param.layers enable layers
	 */
	constructor({
		context,
		fs,
		resolverFactory,
		options,
		associatedObjectForCache,
		layers = false
	***REMOVED***) {
		super();
		this.hooks = Object.freeze({
			/** @type {AsyncSeriesBailHook<[ResolveData], Module | false | void>***REMOVED*** */
			resolve: new AsyncSeriesBailHook(["resolveData"]),
			/** @type {HookMap<AsyncSeriesBailHook<[ResourceDataWithData, ResolveData], true | void>>***REMOVED*** */
			resolveForScheme: new HookMap(
				() => new AsyncSeriesBailHook(["resourceData", "resolveData"])
			),
			/** @type {HookMap<AsyncSeriesBailHook<[ResourceDataWithData, ResolveData], true | void>>***REMOVED*** */
			resolveInScheme: new HookMap(
				() => new AsyncSeriesBailHook(["resourceData", "resolveData"])
			),
			/** @type {AsyncSeriesBailHook<[ResolveData], Module | undefined>***REMOVED*** */
			factorize: new AsyncSeriesBailHook(["resolveData"]),
			/** @type {AsyncSeriesBailHook<[ResolveData], false | void>***REMOVED*** */
			beforeResolve: new AsyncSeriesBailHook(["resolveData"]),
			/** @type {AsyncSeriesBailHook<[ResolveData], false | void>***REMOVED*** */
			afterResolve: new AsyncSeriesBailHook(["resolveData"]),
			/** @type {AsyncSeriesBailHook<[CreateData, ResolveData], Module | void>***REMOVED*** */
			createModule: new AsyncSeriesBailHook(["createData", "resolveData"]),
			/** @type {SyncWaterfallHook<[Module, CreateData, ResolveData]>***REMOVED*** */
			module: new SyncWaterfallHook(["module", "createData", "resolveData"]),
			/** @type {HookMap<SyncBailHook<[ParserOptions], Parser | void>>***REMOVED*** */
			createParser: new HookMap(() => new SyncBailHook(["parserOptions"])),
			/** @type {HookMap<SyncBailHook<[TODO, ParserOptions], void>>***REMOVED*** */
			parser: new HookMap(() => new SyncHook(["parser", "parserOptions"])),
			/** @type {HookMap<SyncBailHook<[GeneratorOptions], Generator | void>>***REMOVED*** */
			createGenerator: new HookMap(
				() => new SyncBailHook(["generatorOptions"])
			),
			/** @type {HookMap<SyncBailHook<[TODO, GeneratorOptions], void>>***REMOVED*** */
			generator: new HookMap(
				() => new SyncHook(["generator", "generatorOptions"])
			),
			/** @type {HookMap<SyncBailHook<[TODO, ResolveData], Module | void>>***REMOVED*** */
			createModuleClass: new HookMap(
				() => new SyncBailHook(["createData", "resolveData"])
			)
		***REMOVED***);
		this.resolverFactory = resolverFactory;
		this.ruleSet = ruleSetCompiler.compile([
			{
				rules: /** @type {RuleSetRules***REMOVED*** */ (options.defaultRules)
			***REMOVED***,
			{
				rules: /** @type {RuleSetRules***REMOVED*** */ (options.rules)
			***REMOVED***
		]);
		this.context = context || "";
		this.fs = fs;
		this._globalParserOptions = options.parser;
		this._globalGeneratorOptions = options.generator;
		/** @type {Map<string, WeakMap<ParserOptions, Parser>>***REMOVED*** */
		this.parserCache = new Map();
		/** @type {Map<string, WeakMap<GeneratorOptions, Generator>>***REMOVED*** */
		this.generatorCache = new Map();
		/** @type {Set<Module>***REMOVED*** */
		this._restoredUnsafeCacheEntries = new Set();

		const cacheParseResource = parseResource.bindCache(
			associatedObjectForCache
		);
		const cachedParseResourceWithoutFragment =
			parseResourceWithoutFragment.bindCache(associatedObjectForCache);
		this._parseResourceWithoutFragment = cachedParseResourceWithoutFragment;

		this.hooks.factorize.tapAsync(
			{
				name: "NormalModuleFactory",
				stage: 100
			***REMOVED***,
			(resolveData, callback) => {
				this.hooks.resolve.callAsync(resolveData, (err, result) => {
					if (err) return callback(err);

					// Ignored
					if (result === false) return callback();

					// direct module
					if (result instanceof Module) return callback(null, result);

					if (typeof result === "object")
						throw new Error(
							`${deprecationChangedHookMessage(
								"resolve",
								this.hooks.resolve
							)***REMOVED*** Returning a Module object will result in this module used as result.`
						);

					this.hooks.afterResolve.callAsync(resolveData, (err, result) => {
						if (err) return callback(err);

						if (typeof result === "object")
							throw new Error(
								deprecationChangedHookMessage(
									"afterResolve",
									this.hooks.afterResolve
								)
							);

						// Ignored
						if (result === false) return callback();

						const createData = resolveData.createData;

						this.hooks.createModule.callAsync(
							createData,
							resolveData,
							(err, createdModule) => {
								if (!createdModule) {
									if (!resolveData.request) {
										return callback(new Error("Empty dependency (no request)"));
									***REMOVED***

									// TODO webpack 6 make it required and move javascript/wasm/asset properties to own module
									createdModule = this.hooks.createModuleClass
										.for(
											/** @type {ModuleSettings***REMOVED*** */
											(createData.settings).type
										)
										.call(createData, resolveData);

									if (!createdModule) {
										createdModule = /** @type {Module***REMOVED*** */ (
											new NormalModule(
												/** @type {NormalModuleCreateData***REMOVED*** */
												(createData)
											)
										);
									***REMOVED***
								***REMOVED***

								createdModule = this.hooks.module.call(
									createdModule,
									createData,
									resolveData
								);

								return callback(null, createdModule);
							***REMOVED***
						);
					***REMOVED***);
				***REMOVED***);
			***REMOVED***
		);
		this.hooks.resolve.tapAsync(
			{
				name: "NormalModuleFactory",
				stage: 100
			***REMOVED***,
			(data, callback) => {
				const {
					contextInfo,
					context,
					dependencies,
					dependencyType,
					request,
					assertions,
					resolveOptions,
					fileDependencies,
					missingDependencies,
					contextDependencies
				***REMOVED*** = data;
				const loaderResolver = this.getResolver("loader");

				/** @type {ResourceData | undefined***REMOVED*** */
				let matchResourceData;
				/** @type {string***REMOVED*** */
				let unresolvedResource;
				/** @type {ParsedLoaderRequest[]***REMOVED*** */
				let elements;
				let noPreAutoLoaders = false;
				let noAutoLoaders = false;
				let noPrePostAutoLoaders = false;

				const contextScheme = getScheme(context);
				/** @type {string | undefined***REMOVED*** */
				let scheme = getScheme(request);

				if (!scheme) {
					/** @type {string***REMOVED*** */
					let requestWithoutMatchResource = request;
					const matchResourceMatch = MATCH_RESOURCE_REGEX.exec(request);
					if (matchResourceMatch) {
						let matchResource = matchResourceMatch[1];
						if (matchResource.charCodeAt(0) === 46) {
							// 46 === ".", 47 === "/"
							const secondChar = matchResource.charCodeAt(1);
							if (
								secondChar === 47 ||
								(secondChar === 46 && matchResource.charCodeAt(2) === 47)
							) {
								// if matchResources startsWith ../ or ./
								matchResource = join(this.fs, context, matchResource);
							***REMOVED***
						***REMOVED***

						matchResourceData = {
							resource: matchResource,
							.../** @type {TODO***REMOVED*** */ (cacheParseResource(matchResource))
						***REMOVED***;
						requestWithoutMatchResource = request.slice(
							matchResourceMatch[0].length
						);
					***REMOVED***

					scheme = getScheme(requestWithoutMatchResource);

					if (!scheme && !contextScheme) {
						const firstChar = requestWithoutMatchResource.charCodeAt(0);
						const secondChar = requestWithoutMatchResource.charCodeAt(1);
						noPreAutoLoaders = firstChar === 45 && secondChar === 33; // startsWith "-!"
						noAutoLoaders = noPreAutoLoaders || firstChar === 33; // startsWith "!"
						noPrePostAutoLoaders = firstChar === 33 && secondChar === 33; // startsWith "!!";
						const rawElements = requestWithoutMatchResource
							.slice(
								noPreAutoLoaders || noPrePostAutoLoaders
									? 2
									: noAutoLoaders
										? 1
										: 0
							)
							.split(/!+/);
						unresolvedResource = /** @type {string***REMOVED*** */ (rawElements.pop());
						elements = rawElements.map(el => {
							const { path, query ***REMOVED*** = cachedParseResourceWithoutFragment(el);
							return {
								loader: path,
								options: query ? query.slice(1) : undefined
							***REMOVED***;
						***REMOVED***);
						scheme = getScheme(unresolvedResource);
					***REMOVED*** else {
						unresolvedResource = requestWithoutMatchResource;
						elements = EMPTY_ELEMENTS;
					***REMOVED***
				***REMOVED*** else {
					unresolvedResource = request;
					elements = EMPTY_ELEMENTS;
				***REMOVED***

				/** @type {ResolveContext***REMOVED*** */
				const resolveContext = {
					fileDependencies,
					missingDependencies,
					contextDependencies
				***REMOVED***;

				/** @type {ResourceDataWithData***REMOVED*** */
				let resourceData;

				/** @type {undefined | LoaderItem[]***REMOVED*** */
				let loaders;

				const continueCallback = needCalls(2, err => {
					if (err) return callback(err);

					// translate option idents
					try {
						for (const item of /** @type {LoaderItem[]***REMOVED*** */ (loaders)) {
							if (typeof item.options === "string" && item.options[0] === "?") {
								const ident = item.options.slice(1);
								if (ident === "[[missing ident]]") {
									throw new Error(
										"No ident is provided by referenced loader. " +
											"When using a function for Rule.use in config you need to " +
											"provide an 'ident' property for referenced loader options."
									);
								***REMOVED***
								item.options = this.ruleSet.references.get(ident);
								if (item.options === undefined) {
									throw new Error(
										"Invalid ident is provided by referenced loader"
									);
								***REMOVED***
								item.ident = ident;
							***REMOVED***
						***REMOVED***
					***REMOVED*** catch (identErr) {
						return callback(/** @type {Error***REMOVED*** */ (identErr));
					***REMOVED***

					if (!resourceData) {
						// ignored
						return callback(
							null,
							/** @type {TODO***REMOVED*** */
							(dependencies[0].createIgnoredModule(context))
						);
					***REMOVED***

					const userRequest =
						(matchResourceData !== undefined
							? `${matchResourceData.resource***REMOVED***!=!`
							: "") +
						stringifyLoadersAndResource(
							/** @type {LoaderItem[]***REMOVED*** */ (loaders),
							resourceData.resource
						);

					/** @type {ModuleSettings***REMOVED*** */
					const settings = {***REMOVED***;
					const useLoadersPost = [];
					const useLoaders = [];
					const useLoadersPre = [];

					// handle .webpack[] suffix
					let resource;
					let match;
					if (
						matchResourceData &&
						typeof (resource = matchResourceData.resource) === "string" &&
						(match = /\.webpack\[([^\]]+)\]$/.exec(resource))
					) {
						settings.type = match[1];
						matchResourceData.resource = matchResourceData.resource.slice(
							0,
							-settings.type.length - 10
						);
					***REMOVED*** else {
						settings.type = JAVASCRIPT_MODULE_TYPE_AUTO;
						const resourceDataForRules = matchResourceData || resourceData;
						const result = this.ruleSet.exec({
							resource: resourceDataForRules.path,
							realResource: resourceData.path,
							resourceQuery: resourceDataForRules.query,
							resourceFragment: resourceDataForRules.fragment,
							scheme,
							assertions,
							mimetype: matchResourceData
								? ""
								: resourceData.data.mimetype || "",
							dependency: dependencyType,
							descriptionData: matchResourceData
								? undefined
								: resourceData.data.descriptionFileData,
							issuer: contextInfo.issuer,
							compiler: contextInfo.compiler,
							issuerLayer: contextInfo.issuerLayer || ""
						***REMOVED***);
						for (const r of result) {
							// https://github.com/webpack/webpack/issues/16466
							// if a request exists PrePostAutoLoaders, should disable modifying Rule.type
							if (r.type === "type" && noPrePostAutoLoaders) {
								continue;
							***REMOVED***
							if (r.type === "use") {
								if (!noAutoLoaders && !noPrePostAutoLoaders) {
									useLoaders.push(r.value);
								***REMOVED***
							***REMOVED*** else if (r.type === "use-post") {
								if (!noPrePostAutoLoaders) {
									useLoadersPost.push(r.value);
								***REMOVED***
							***REMOVED*** else if (r.type === "use-pre") {
								if (!noPreAutoLoaders && !noPrePostAutoLoaders) {
									useLoadersPre.push(r.value);
								***REMOVED***
							***REMOVED*** else if (
								typeof r.value === "object" &&
								r.value !== null &&
								typeof settings[
									/** @type {keyof ModuleSettings***REMOVED*** */ (r.type)
								] === "object" &&
								settings[/** @type {keyof ModuleSettings***REMOVED*** */ (r.type)] !== null
							) {
								const type = /** @type {keyof ModuleSettings***REMOVED*** */ (r.type);
								/** @type {TODO***REMOVED*** */
								(settings)[type] = cachedCleverMerge(settings[type], r.value);
							***REMOVED*** else {
								const type = /** @type {keyof ModuleSettings***REMOVED*** */ (r.type);
								/** @type {TODO***REMOVED*** */
								(settings)[type] = r.value;
							***REMOVED***
						***REMOVED***
					***REMOVED***

					/** @type {undefined | LoaderItem[]***REMOVED*** */
					let postLoaders;
					/** @type {undefined | LoaderItem[]***REMOVED*** */
					let normalLoaders;
					/** @type {undefined | LoaderItem[]***REMOVED*** */
					let preLoaders;

					const continueCallback = needCalls(3, err => {
						if (err) {
							return callback(err);
						***REMOVED***
						const allLoaders = /** @type {LoaderItem[]***REMOVED*** */ (postLoaders);
						if (matchResourceData === undefined) {
							for (const loader of /** @type {LoaderItem[]***REMOVED*** */ (loaders))
								allLoaders.push(loader);
							for (const loader of /** @type {LoaderItem[]***REMOVED*** */ (normalLoaders))
								allLoaders.push(loader);
						***REMOVED*** else {
							for (const loader of /** @type {LoaderItem[]***REMOVED*** */ (normalLoaders))
								allLoaders.push(loader);
							for (const loader of /** @type {LoaderItem[]***REMOVED*** */ (loaders))
								allLoaders.push(loader);
						***REMOVED***
						for (const loader of /** @type {LoaderItem[]***REMOVED*** */ (preLoaders))
							allLoaders.push(loader);
						const type = /** @type {string***REMOVED*** */ (settings.type);
						const resolveOptions = settings.resolve;
						const layer = settings.layer;
						if (layer !== undefined && !layers) {
							return callback(
								new Error(
									"'Rule.layer' is only allowed when 'experiments.layers' is enabled"
								)
							);
						***REMOVED***
						try {
							Object.assign(data.createData, {
								layer:
									layer === undefined ? contextInfo.issuerLayer || null : layer,
								request: stringifyLoadersAndResource(
									allLoaders,
									resourceData.resource
								),
								userRequest,
								rawRequest: request,
								loaders: allLoaders,
								resource: resourceData.resource,
								context:
									resourceData.context || getContext(resourceData.resource),
								matchResource: matchResourceData
									? matchResourceData.resource
									: undefined,
								resourceResolveData: resourceData.data,
								settings,
								type,
								parser: this.getParser(type, settings.parser),
								parserOptions: settings.parser,
								generator: this.getGenerator(type, settings.generator),
								generatorOptions: settings.generator,
								resolveOptions
							***REMOVED***);
						***REMOVED*** catch (createDataErr) {
							return callback(/** @type {Error***REMOVED*** */ (createDataErr));
						***REMOVED***
						callback();
					***REMOVED***);
					this.resolveRequestArray(
						contextInfo,
						this.context,
						useLoadersPost,
						loaderResolver,
						resolveContext,
						(err, result) => {
							postLoaders = result;
							continueCallback(err);
						***REMOVED***
					);
					this.resolveRequestArray(
						contextInfo,
						this.context,
						useLoaders,
						loaderResolver,
						resolveContext,
						(err, result) => {
							normalLoaders = result;
							continueCallback(err);
						***REMOVED***
					);
					this.resolveRequestArray(
						contextInfo,
						this.context,
						useLoadersPre,
						loaderResolver,
						resolveContext,
						(err, result) => {
							preLoaders = result;
							continueCallback(err);
						***REMOVED***
					);
				***REMOVED***);

				this.resolveRequestArray(
					contextInfo,
					contextScheme ? this.context : context,
					/** @type {LoaderItem[]***REMOVED*** */ (elements),
					loaderResolver,
					resolveContext,
					(err, result) => {
						if (err) return continueCallback(err);
						loaders = result;
						continueCallback();
					***REMOVED***
				);

				/**
				 * @param {string***REMOVED*** context context
				 */
				const defaultResolve = context => {
					if (/^($|\?)/.test(unresolvedResource)) {
						resourceData = {
							resource: unresolvedResource,
							data: {***REMOVED***,
							.../** @type {TODO***REMOVED*** */ (cacheParseResource(unresolvedResource))
						***REMOVED***;
						continueCallback();
					***REMOVED***

					// resource without scheme and with path
					else {
						const normalResolver = this.getResolver(
							"normal",
							dependencyType
								? cachedSetProperty(
										resolveOptions || EMPTY_RESOLVE_OPTIONS,
										"dependencyType",
										dependencyType
									)
								: resolveOptions
						);
						this.resolveResource(
							contextInfo,
							context,
							unresolvedResource,
							normalResolver,
							resolveContext,
							(err, _resolvedResource, resolvedResourceResolveData) => {
								if (err) return continueCallback(err);
								if (_resolvedResource !== false) {
									const resolvedResource =
										/** @type {string***REMOVED*** */
										(_resolvedResource);
									resourceData = {
										resource: resolvedResource,
										data:
											/** @type {ResolveRequest***REMOVED*** */
											(resolvedResourceResolveData),
										.../** @type {TODO***REMOVED*** */
										(cacheParseResource(resolvedResource))
									***REMOVED***;
								***REMOVED***
								continueCallback();
							***REMOVED***
						);
					***REMOVED***
				***REMOVED***;

				// resource with scheme
				if (scheme) {
					resourceData = {
						resource: unresolvedResource,
						data: {***REMOVED***,
						path: undefined,
						query: undefined,
						fragment: undefined,
						context: undefined
					***REMOVED***;
					this.hooks.resolveForScheme
						.for(scheme)
						.callAsync(resourceData, data, err => {
							if (err) return continueCallback(err);
							continueCallback();
						***REMOVED***);
				***REMOVED***

				// resource within scheme
				else if (contextScheme) {
					resourceData = {
						resource: unresolvedResource,
						data: {***REMOVED***,
						path: undefined,
						query: undefined,
						fragment: undefined,
						context: undefined
					***REMOVED***;
					this.hooks.resolveInScheme
						.for(contextScheme)
						.callAsync(resourceData, data, (err, handled) => {
							if (err) return continueCallback(err);
							if (!handled) return defaultResolve(this.context);
							continueCallback();
						***REMOVED***);
				***REMOVED***

				// resource without scheme and without path
				else defaultResolve(context);
			***REMOVED***
		);
	***REMOVED***

	cleanupForCache() {
		for (const module of this._restoredUnsafeCacheEntries) {
			ChunkGraph.clearChunkGraphForModule(module);
			ModuleGraph.clearModuleGraphForModule(module);
			module.cleanupForCache();
		***REMOVED***
	***REMOVED***

	/**
	 * @param {ModuleFactoryCreateData***REMOVED*** data data object
	 * @param {ModuleFactoryCallback***REMOVED*** callback callback
	 * @returns {void***REMOVED***
	 */
	create(data, callback) {
		const dependencies = /** @type {ModuleDependency[]***REMOVED*** */ (data.dependencies);
		const context = data.context || this.context;
		const resolveOptions = data.resolveOptions || EMPTY_RESOLVE_OPTIONS;
		const dependency = dependencies[0];
		const request = dependency.request;
		const assertions = dependency.assertions;
		const dependencyType = dependency.category || "";
		const contextInfo = data.contextInfo;
		const fileDependencies = new LazySet();
		const missingDependencies = new LazySet();
		const contextDependencies = new LazySet();
		/** @type {ResolveData***REMOVED*** */
		const resolveData = {
			contextInfo,
			resolveOptions,
			context,
			request,
			assertions,
			dependencies,
			dependencyType,
			fileDependencies,
			missingDependencies,
			contextDependencies,
			createData: {***REMOVED***,
			cacheable: true
		***REMOVED***;
		this.hooks.beforeResolve.callAsync(resolveData, (err, result) => {
			if (err) {
				return callback(err, {
					fileDependencies,
					missingDependencies,
					contextDependencies,
					cacheable: false
				***REMOVED***);
			***REMOVED***

			// Ignored
			if (result === false) {
				/** @type {ModuleFactoryResult***REMOVED*** * */
				const factoryResult = {
					fileDependencies,
					missingDependencies,
					contextDependencies,
					cacheable: resolveData.cacheable
				***REMOVED***;

				if (resolveData.ignoredModule) {
					factoryResult.module = resolveData.ignoredModule;
				***REMOVED***

				return callback(null, factoryResult);
			***REMOVED***

			if (typeof result === "object")
				throw new Error(
					deprecationChangedHookMessage(
						"beforeResolve",
						this.hooks.beforeResolve
					)
				);

			this.hooks.factorize.callAsync(resolveData, (err, module) => {
				if (err) {
					return callback(err, {
						fileDependencies,
						missingDependencies,
						contextDependencies,
						cacheable: false
					***REMOVED***);
				***REMOVED***

				/** @type {ModuleFactoryResult***REMOVED*** * */
				const factoryResult = {
					module,
					fileDependencies,
					missingDependencies,
					contextDependencies,
					cacheable: resolveData.cacheable
				***REMOVED***;

				callback(null, factoryResult);
			***REMOVED***);
		***REMOVED***);
	***REMOVED***

	/**
	 * @param {ModuleFactoryCreateDataContextInfo***REMOVED*** contextInfo context info
	 * @param {string***REMOVED*** context context
	 * @param {string***REMOVED*** unresolvedResource unresolved resource
	 * @param {ResolverWithOptions***REMOVED*** resolver resolver
	 * @param {ResolveContext***REMOVED*** resolveContext resolver context
	 * @param {(err: null | Error, res?: string | false, req?: ResolveRequest) => void***REMOVED*** callback callback
	 */
	resolveResource(
		contextInfo,
		context,
		unresolvedResource,
		resolver,
		resolveContext,
		callback
	) {
		resolver.resolve(
			contextInfo,
			context,
			unresolvedResource,
			resolveContext,
			(err, resolvedResource, resolvedResourceResolveData) => {
				if (err) {
					return this._resolveResourceErrorHints(
						err,
						contextInfo,
						context,
						unresolvedResource,
						resolver,
						resolveContext,
						(err2, hints) => {
							if (err2) {
								err.message += `
A fatal error happened during resolving additional hints for this error: ${err2.message***REMOVED***`;
								err.stack += `

A fatal error happened during resolving additional hints for this error:
${err2.stack***REMOVED***`;
								return callback(err);
							***REMOVED***
							if (hints && hints.length > 0) {
								err.message += `
${hints.join("\n\n")***REMOVED***`;
							***REMOVED***

							// Check if the extension is missing a leading dot (e.g. "js" instead of ".js")
							let appendResolveExtensionsHint = false;
							const specifiedExtensions = Array.from(
								resolver.options.extensions
							);
							const expectedExtensions = specifiedExtensions.map(extension => {
								if (LEADING_DOT_EXTENSION_REGEX.test(extension)) {
									appendResolveExtensionsHint = true;
									return `.${extension***REMOVED***`;
								***REMOVED***
								return extension;
							***REMOVED***);
							if (appendResolveExtensionsHint) {
								err.message += `\nDid you miss the leading dot in 'resolve.extensions'? Did you mean '${JSON.stringify(
									expectedExtensions
								)***REMOVED***' instead of '${JSON.stringify(specifiedExtensions)***REMOVED***'?`;
							***REMOVED***

							callback(err);
						***REMOVED***
					);
				***REMOVED***
				callback(err, resolvedResource, resolvedResourceResolveData);
			***REMOVED***
		);
	***REMOVED***

	/**
	 * @param {Error***REMOVED*** error error
	 * @param {ModuleFactoryCreateDataContextInfo***REMOVED*** contextInfo context info
	 * @param {string***REMOVED*** context context
	 * @param {string***REMOVED*** unresolvedResource unresolved resource
	 * @param {ResolverWithOptions***REMOVED*** resolver resolver
	 * @param {ResolveContext***REMOVED*** resolveContext resolver context
	 * @param {Callback<string[]>***REMOVED*** callback callback
	 * @private
	 */
	_resolveResourceErrorHints(
		error,
		contextInfo,
		context,
		unresolvedResource,
		resolver,
		resolveContext,
		callback
	) {
		asyncLib.parallel(
			[
				callback => {
					if (!resolver.options.fullySpecified) return callback();
					resolver
						.withOptions({
							fullySpecified: false
						***REMOVED***)
						.resolve(
							contextInfo,
							context,
							unresolvedResource,
							resolveContext,
							(err, resolvedResource) => {
								if (!err && resolvedResource) {
									const resource = parseResource(resolvedResource).path.replace(
										/^.*[\\/]/,
										""
									);
									return callback(
										null,
										`Did you mean '${resource***REMOVED***'?
BREAKING CHANGE: The request '${unresolvedResource***REMOVED***' failed to resolve only because it was resolved as fully specified
(probably because the origin is strict EcmaScript Module, e. g. a module with javascript mimetype, a '*.mjs' file, or a '*.js' file where the package.json contains '"type": "module"').
The extension in the request is mandatory for it to be fully specified.
Add the extension to the request.`
									);
								***REMOVED***
								callback();
							***REMOVED***
						);
				***REMOVED***,
				callback => {
					if (!resolver.options.enforceExtension) return callback();
					resolver
						.withOptions({
							enforceExtension: false,
							extensions: []
						***REMOVED***)
						.resolve(
							contextInfo,
							context,
							unresolvedResource,
							resolveContext,
							(err, resolvedResource) => {
								if (!err && resolvedResource) {
									let hint = "";
									const match = /(\.[^.]+)(\?|$)/.exec(unresolvedResource);
									if (match) {
										const fixedRequest = unresolvedResource.replace(
											/(\.[^.]+)(\?|$)/,
											"$2"
										);
										hint = resolver.options.extensions.has(match[1])
											? `Did you mean '${fixedRequest***REMOVED***'?`
											: `Did you mean '${fixedRequest***REMOVED***'? Also note that '${match[1]***REMOVED***' is not in 'resolve.extensions' yet and need to be added for this to work?`;
									***REMOVED*** else {
										hint =
											"Did you mean to omit the extension or to remove 'resolve.enforceExtension'?";
									***REMOVED***
									return callback(
										null,
										`The request '${unresolvedResource***REMOVED***' failed to resolve only because 'resolve.enforceExtension' was specified.
${hint***REMOVED***
Including the extension in the request is no longer possible. Did you mean to enforce including the extension in requests with 'resolve.extensions: []' instead?`
									);
								***REMOVED***
								callback();
							***REMOVED***
						);
				***REMOVED***,
				callback => {
					if (
						/^\.\.?\//.test(unresolvedResource) ||
						resolver.options.preferRelative
					) {
						return callback();
					***REMOVED***
					resolver.resolve(
						contextInfo,
						context,
						`./${unresolvedResource***REMOVED***`,
						resolveContext,
						(err, resolvedResource) => {
							if (err || !resolvedResource) return callback();
							const moduleDirectories = resolver.options.modules
								.map(m => (Array.isArray(m) ? m.join(", ") : m))
								.join(", ");
							callback(
								null,
								`Did you mean './${unresolvedResource***REMOVED***'?
Requests that should resolve in the current directory need to start with './'.
Requests that start with a name are treated as module requests and resolve within module directories (${moduleDirectories***REMOVED***).
If changing the source code is not an option there is also a resolve options called 'preferRelative' which tries to resolve these kind of requests in the current directory too.`
							);
						***REMOVED***
					);
				***REMOVED***
			],
			(err, hints) => {
				if (err) return callback(err);
				callback(null, /** @type {string[]***REMOVED*** */ (hints).filter(Boolean));
			***REMOVED***
		);
	***REMOVED***

	/**
	 * @param {ModuleFactoryCreateDataContextInfo***REMOVED*** contextInfo context info
	 * @param {string***REMOVED*** context context
	 * @param {LoaderItem[]***REMOVED*** array array
	 * @param {ResolverWithOptions***REMOVED*** resolver resolver
	 * @param {ResolveContext***REMOVED*** resolveContext resolve context
	 * @param {Callback<LoaderItem[]>***REMOVED*** callback callback
	 * @returns {void***REMOVED*** result
	 */
	resolveRequestArray(
		contextInfo,
		context,
		array,
		resolver,
		resolveContext,
		callback
	) {
		// LoaderItem
		if (array.length === 0) return callback(null, array);
		asyncLib.map(
			array,
			(item, callback) => {
				resolver.resolve(
					contextInfo,
					context,
					item.loader,
					resolveContext,
					(err, result, resolveRequest) => {
						if (
							err &&
							/^[^/]*$/.test(item.loader) &&
							!item.loader.endsWith("-loader")
						) {
							return resolver.resolve(
								contextInfo,
								context,
								`${item.loader***REMOVED***-loader`,
								resolveContext,
								err2 => {
									if (!err2) {
										err.message =
											`${err.message***REMOVED***\n` +
											"BREAKING CHANGE: It's no longer allowed to omit the '-loader' suffix when using loaders.\n" +
											`                 You need to specify '${item.loader***REMOVED***-loader' instead of '${item.loader***REMOVED***',\n` +
											"                 see https://webpack.js.org/migrate/3/#automatic-loader-module-name-extension-removed";
									***REMOVED***
									callback(err);
								***REMOVED***
							);
						***REMOVED***
						if (err) return callback(err);

						const parsedResult = this._parseResourceWithoutFragment(
							/** @type {string***REMOVED*** */ (result)
						);

						const type = /\.mjs$/i.test(parsedResult.path)
							? "module"
							: /\.cjs$/i.test(parsedResult.path)
								? "commonjs"
								: /** @type {ResolveRequest***REMOVED*** */
									(resolveRequest).descriptionFileData === undefined
									? undefined
									: /** @type {ResolveRequest***REMOVED*** */
										(resolveRequest).descriptionFileData.type;
						const resolved = {
							loader: parsedResult.path,
							type,
							options:
								item.options === undefined
									? parsedResult.query
										? parsedResult.query.slice(1)
										: undefined
									: item.options,
							ident:
								item.options === undefined
									? undefined
									: /** @type {string***REMOVED*** */ (item.ident)
						***REMOVED***;

						return callback(null, /** @type {LoaderItem***REMOVED*** */ (resolved));
					***REMOVED***
				);
			***REMOVED***,
			/** @type {Callback<TODO>***REMOVED*** */ (callback)
		);
	***REMOVED***

	/**
	 * @param {string***REMOVED*** type type
	 * @param {ParserOptions***REMOVED*** parserOptions parser options
	 * @returns {Parser***REMOVED*** parser
	 */
	getParser(type, parserOptions = EMPTY_PARSER_OPTIONS) {
		let cache = this.parserCache.get(type);

		if (cache === undefined) {
			cache = new WeakMap();
			this.parserCache.set(type, cache);
		***REMOVED***

		let parser = cache.get(parserOptions);

		if (parser === undefined) {
			parser = this.createParser(type, parserOptions);
			cache.set(parserOptions, parser);
		***REMOVED***

		return parser;
	***REMOVED***

	/**
	 * @param {string***REMOVED*** type type
	 * @param {ParserOptions***REMOVED*** parserOptions parser options
	 * @returns {Parser***REMOVED*** parser
	 */
	createParser(type, parserOptions = {***REMOVED***) {
		parserOptions = mergeGlobalOptions(
			this._globalParserOptions,
			type,
			parserOptions
		);
		const parser = this.hooks.createParser.for(type).call(parserOptions);
		if (!parser) {
			throw new Error(`No parser registered for ${type***REMOVED***`);
		***REMOVED***
		this.hooks.parser.for(type).call(parser, parserOptions);
		return parser;
	***REMOVED***

	/**
	 * @param {string***REMOVED*** type type of generator
	 * @param {GeneratorOptions***REMOVED*** generatorOptions generator options
	 * @returns {Generator***REMOVED*** generator
	 */
	getGenerator(type, generatorOptions = EMPTY_GENERATOR_OPTIONS) {
		let cache = this.generatorCache.get(type);

		if (cache === undefined) {
			cache = new WeakMap();
			this.generatorCache.set(type, cache);
		***REMOVED***

		let generator = cache.get(generatorOptions);

		if (generator === undefined) {
			generator = this.createGenerator(type, generatorOptions);
			cache.set(generatorOptions, generator);
		***REMOVED***

		return generator;
	***REMOVED***

	/**
	 * @param {string***REMOVED*** type type of generator
	 * @param {GeneratorOptions***REMOVED*** generatorOptions generator options
	 * @returns {Generator***REMOVED*** generator
	 */
	createGenerator(type, generatorOptions = {***REMOVED***) {
		generatorOptions = mergeGlobalOptions(
			this._globalGeneratorOptions,
			type,
			generatorOptions
		);
		const generator = this.hooks.createGenerator
			.for(type)
			.call(generatorOptions);
		if (!generator) {
			throw new Error(`No generator registered for ${type***REMOVED***`);
		***REMOVED***
		this.hooks.generator.for(type).call(generator, generatorOptions);
		return generator;
	***REMOVED***

	/**
	 * @param {Parameters<ResolverFactory["get"]>[0]***REMOVED*** type type of resolver
	 * @param {Parameters<ResolverFactory["get"]>[1]=***REMOVED*** resolveOptions options
	 * @returns {ReturnType<ResolverFactory["get"]>***REMOVED*** the resolver
	 */
	getResolver(type, resolveOptions) {
		return this.resolverFactory.get(type, resolveOptions);
	***REMOVED***
***REMOVED***

module.exports = NormalModuleFactory;
