// webpack/lib/NormalModule.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const parseJson = require("json-parse-even-better-errors");
const { getContext, runLoaders ***REMOVED*** = require("loader-runner");
const querystring = require("querystring");
const {
	HookMap,
	SyncHook,
	SyncWaterfallHook,
	AsyncSeriesBailHook
***REMOVED*** = require("tapable");
const {
	CachedSource,
	OriginalSource,
	RawSource,
	SourceMapSource
***REMOVED*** = require("webpack-sources");
const Compilation = require("./Compilation");
const HookWebpackError = require("./HookWebpackError");
const Module = require("./Module");
const ModuleBuildError = require("./ModuleBuildError");
const ModuleError = require("./ModuleError");
const ModuleGraphConnection = require("./ModuleGraphConnection");
const ModuleParseError = require("./ModuleParseError");
const { JAVASCRIPT_MODULE_TYPE_AUTO ***REMOVED*** = require("./ModuleTypeConstants");
const ModuleWarning = require("./ModuleWarning");
const RuntimeGlobals = require("./RuntimeGlobals");
const UnhandledSchemeError = require("./UnhandledSchemeError");
const WebpackError = require("./WebpackError");
const formatLocation = require("./formatLocation");
const LazySet = require("./util/LazySet");
const { isSubset ***REMOVED*** = require("./util/SetHelpers");
const { getScheme ***REMOVED*** = require("./util/URLAbsoluteSpecifier");
const {
	compareLocations,
	concatComparators,
	compareSelect,
	keepOriginalOrder
***REMOVED*** = require("./util/comparators");
const createHash = require("./util/createHash");
const { createFakeHook ***REMOVED*** = require("./util/deprecation");
const { join ***REMOVED*** = require("./util/fs");
const {
	contextify,
	absolutify,
	makePathsRelative
***REMOVED*** = require("./util/identifier");
const makeSerializable = require("./util/makeSerializable");
const memoize = require("./util/memoize");

/** @typedef {import("webpack-sources").Source***REMOVED*** Source */
/** @typedef {import("../declarations/WebpackOptions").Mode***REMOVED*** Mode */
/** @typedef {import("../declarations/WebpackOptions").ResolveOptions***REMOVED*** ResolveOptions */
/** @typedef {import("../declarations/WebpackOptions").WebpackOptionsNormalized***REMOVED*** WebpackOptions */
/** @typedef {import("../declarations/WebpackOptions").NoParse***REMOVED*** NoParse */
/** @typedef {import("./ChunkGraph")***REMOVED*** ChunkGraph */
/** @typedef {import("./Compiler")***REMOVED*** Compiler */
/** @typedef {import("./Dependency").UpdateHashContext***REMOVED*** UpdateHashContext */
/** @typedef {import("./DependencyTemplates")***REMOVED*** DependencyTemplates */
/** @typedef {import("./Generator")***REMOVED*** Generator */
/** @typedef {import("./Generator").GenerateErrorFn***REMOVED*** GenerateErrorFn */
/** @typedef {import("./Module").BuildInfo***REMOVED*** BuildInfo */
/** @typedef {import("./Module").BuildMeta***REMOVED*** BuildMeta */
/** @typedef {import("./Module").CodeGenerationContext***REMOVED*** CodeGenerationContext */
/** @typedef {import("./Module").CodeGenerationResult***REMOVED*** CodeGenerationResult */
/** @typedef {import("./Module").ConcatenationBailoutReasonContext***REMOVED*** ConcatenationBailoutReasonContext */
/** @typedef {import("./Module").KnownBuildInfo***REMOVED*** KnownBuildInfo */
/** @typedef {import("./Module").LibIdentOptions***REMOVED*** LibIdentOptions */
/** @typedef {import("./Module").NeedBuildContext***REMOVED*** NeedBuildContext */
/** @typedef {import("./Module").NeedBuildCallback***REMOVED*** NeedBuildCallback */
/** @typedef {import("./Module").BuildCallback***REMOVED*** BuildCallback */
/** @typedef {import("./Generator").SourceTypes***REMOVED*** SourceTypes */
/** @typedef {import("./Module").UnsafeCacheData***REMOVED*** UnsafeCacheData */
/** @typedef {import("./ModuleGraph")***REMOVED*** ModuleGraph */
/** @typedef {import("./ModuleGraphConnection").ConnectionState***REMOVED*** ConnectionState */
/** @typedef {import("./ModuleTypeConstants").JavaScriptModuleTypes***REMOVED*** JavaScriptModuleTypes */
/** @typedef {import("./NormalModuleFactory")***REMOVED*** NormalModuleFactory */
/** @typedef {import("./NormalModuleFactory").ResourceDataWithData***REMOVED*** ResourceDataWithData */
/** @typedef {import("./Parser")***REMOVED*** Parser */
/** @typedef {import("./Parser").PreparsedAst***REMOVED*** PreparsedAst */
/** @typedef {import("./RequestShortener")***REMOVED*** RequestShortener */
/** @typedef {import("./ResolverFactory").ResolveContext***REMOVED*** ResolveContext */
/** @typedef {import("./ResolverFactory").ResolverWithOptions***REMOVED*** ResolverWithOptions */
/** @typedef {import("./RuntimeTemplate")***REMOVED*** RuntimeTemplate */
/** @typedef {import("./logging/Logger").Logger***REMOVED*** WebpackLogger */
/** @typedef {import("./serialization/ObjectMiddleware").ObjectDeserializerContext***REMOVED*** ObjectDeserializerContext */
/** @typedef {import("./serialization/ObjectMiddleware").ObjectSerializerContext***REMOVED*** ObjectSerializerContext */
/** @typedef {import("./util/Hash")***REMOVED*** Hash */
/** @typedef {import("./util/fs").InputFileSystem***REMOVED*** InputFileSystem */
/** @typedef {import("./util/runtime").RuntimeSpec***REMOVED*** RuntimeSpec */
/** @typedef {import("../declarations/WebpackOptions").HashFunction***REMOVED*** HashFunction */
/** @typedef {import("./util/identifier").AssociatedObjectForCache***REMOVED*** AssociatedObjectForCache */
/**
 * @template T
 * @typedef {import("./util/deprecation").FakeHook<T>***REMOVED*** FakeHook
 */

/** @typedef {{ [k: string]: EXPECTED_ANY ***REMOVED******REMOVED*** ParserOptions */
/** @typedef {{ [k: string]: EXPECTED_ANY ***REMOVED******REMOVED*** GeneratorOptions */

/**
 * @template T
 * @typedef {import("../declarations/LoaderContext").LoaderContext<T>***REMOVED*** LoaderContext
 */

/**
 * @template T
 * @typedef {import("../declarations/LoaderContext").NormalModuleLoaderContext<T>***REMOVED*** NormalModuleLoaderContext
 */

/**
 * @typedef {object***REMOVED*** SourceMap
 * @property {number***REMOVED*** version
 * @property {string[]***REMOVED*** sources
 * @property {string***REMOVED*** mappings
 * @property {string=***REMOVED*** file
 * @property {string=***REMOVED*** sourceRoot
 * @property {string[]=***REMOVED*** sourcesContent
 * @property {string[]=***REMOVED*** names
 * @property {string=***REMOVED*** debugId
 */

const getInvalidDependenciesModuleWarning = memoize(() =>
	require("./InvalidDependenciesModuleWarning")
);
const getValidate = memoize(() => require("schema-utils").validate);

const ABSOLUTE_PATH_REGEX = /^([a-zA-Z]:\\|\\\\|\/)/;

/**
 * @typedef {object***REMOVED*** LoaderItem
 * @property {string***REMOVED*** loader
 * @property {string | null | undefined | Record<string, EXPECTED_ANY>***REMOVED*** options
 * @property {string?***REMOVED*** ident
 * @property {string?***REMOVED*** type
 */

/**
 * @param {string***REMOVED*** context absolute context path
 * @param {string***REMOVED*** source a source path
 * @param {AssociatedObjectForCache=***REMOVED*** associatedObjectForCache an object to which the cache will be attached
 * @returns {string***REMOVED*** new source path
 */
const contextifySourceUrl = (context, source, associatedObjectForCache) => {
	if (source.startsWith("webpack://")) return source;
	return `webpack://${makePathsRelative(
		context,
		source,
		associatedObjectForCache
	)***REMOVED***`;
***REMOVED***;

/**
 * @param {string***REMOVED*** context absolute context path
 * @param {SourceMap***REMOVED*** sourceMap a source map
 * @param {AssociatedObjectForCache=***REMOVED*** associatedObjectForCache an object to which the cache will be attached
 * @returns {SourceMap***REMOVED*** new source map
 */
const contextifySourceMap = (context, sourceMap, associatedObjectForCache) => {
	if (!Array.isArray(sourceMap.sources)) return sourceMap;
	const { sourceRoot ***REMOVED*** = sourceMap;
	/** @type {(source: string) => string***REMOVED*** */
	const mapper = !sourceRoot
		? source => source
		: sourceRoot.endsWith("/")
			? source =>
					source.startsWith("/")
						? `${sourceRoot.slice(0, -1)***REMOVED***${source***REMOVED***`
						: `${sourceRoot***REMOVED***${source***REMOVED***`
			: source =>
					source.startsWith("/")
						? `${sourceRoot***REMOVED***${source***REMOVED***`
						: `${sourceRoot***REMOVED***/${source***REMOVED***`;
	const newSources = sourceMap.sources.map(source =>
		contextifySourceUrl(context, mapper(source), associatedObjectForCache)
	);
	return {
		...sourceMap,
		file: "x",
		sourceRoot: undefined,
		sources: newSources
	***REMOVED***;
***REMOVED***;

/**
 * @param {string | Buffer***REMOVED*** input the input
 * @returns {string***REMOVED*** the converted string
 */
const asString = input => {
	if (Buffer.isBuffer(input)) {
		return input.toString("utf-8");
	***REMOVED***
	return input;
***REMOVED***;

/**
 * @param {string | Buffer***REMOVED*** input the input
 * @returns {Buffer***REMOVED*** the converted buffer
 */
const asBuffer = input => {
	if (!Buffer.isBuffer(input)) {
		return Buffer.from(input, "utf-8");
	***REMOVED***
	return input;
***REMOVED***;

class NonErrorEmittedError extends WebpackError {
	/**
	 * @param {EXPECTED_ANY***REMOVED*** error value which is not an instance of Error
	 */
	constructor(error) {
		super();

		this.name = "NonErrorEmittedError";
		this.message = `(Emitted value instead of an instance of Error) ${error***REMOVED***`;
	***REMOVED***
***REMOVED***

makeSerializable(
	NonErrorEmittedError,
	"webpack/lib/NormalModule",
	"NonErrorEmittedError"
);

/** @typedef {[string | Buffer, string | SourceMapSource, PreparsedAst]***REMOVED***  Result */

/**
 * @typedef {object***REMOVED*** NormalModuleCompilationHooks
 * @property {SyncHook<[LoaderContext<EXPECTED_ANY>, NormalModule]>***REMOVED*** loader
 * @property {SyncHook<[LoaderItem[], NormalModule, LoaderContext<EXPECTED_ANY>]>***REMOVED*** beforeLoaders
 * @property {SyncHook<[NormalModule]>***REMOVED*** beforeParse
 * @property {SyncHook<[NormalModule]>***REMOVED*** beforeSnapshot
 * @property {HookMap<FakeHook<AsyncSeriesBailHook<[string, NormalModule], string | Buffer | null>>>***REMOVED*** readResourceForScheme
 * @property {HookMap<AsyncSeriesBailHook<[LoaderContext<EXPECTED_ANY>], string | Buffer | null>>***REMOVED*** readResource
 * @property {SyncWaterfallHook<[Result, NormalModule]>***REMOVED*** processResult
 * @property {AsyncSeriesBailHook<[NormalModule, NeedBuildContext], boolean>***REMOVED*** needBuild
 */

/**
 * @typedef {object***REMOVED*** NormalModuleCreateData
 * @property {string=***REMOVED*** layer an optional layer in which the module is
 * @property {JavaScriptModuleTypes | ""***REMOVED*** type module type. When deserializing, this is set to an empty string "".
 * @property {string***REMOVED*** request request string
 * @property {string***REMOVED*** userRequest request intended by user (without loaders from config)
 * @property {string***REMOVED*** rawRequest request without resolving
 * @property {LoaderItem[]***REMOVED*** loaders list of loaders
 * @property {string***REMOVED*** resource path + query of the real resource
 * @property {TODO=***REMOVED*** resourceResolveData resource resolve data
 * @property {string***REMOVED*** context context directory for resolving
 * @property {string=***REMOVED*** matchResource path + query of the matched resource (virtual)
 * @property {Parser***REMOVED*** parser the parser used
 * @property {ParserOptions=***REMOVED*** parserOptions the options of the parser used
 * @property {Generator***REMOVED*** generator the generator used
 * @property {GeneratorOptions=***REMOVED*** generatorOptions the options of the generator used
 * @property {ResolveOptions=***REMOVED*** resolveOptions options used for resolving requests from this module
 */

/** @type {WeakMap<Compilation, NormalModuleCompilationHooks>***REMOVED*** */
const compilationHooksMap = new WeakMap();

/** @typedef {Map<string, EXPECTED_ANY>***REMOVED*** CodeGeneratorData */

class NormalModule extends Module {
	/**
	 * @param {Compilation***REMOVED*** compilation the compilation
	 * @returns {NormalModuleCompilationHooks***REMOVED*** the attached hooks
	 */
	static getCompilationHooks(compilation) {
		if (!(compilation instanceof Compilation)) {
			throw new TypeError(
				"The 'compilation' argument must be an instance of Compilation"
			);
		***REMOVED***
		let hooks = compilationHooksMap.get(compilation);
		if (hooks === undefined) {
			hooks = {
				loader: new SyncHook(["loaderContext", "module"]),
				beforeLoaders: new SyncHook(["loaders", "module", "loaderContext"]),
				beforeParse: new SyncHook(["module"]),
				beforeSnapshot: new SyncHook(["module"]),
				// TODO webpack 6 deprecate
				readResourceForScheme: new HookMap(scheme => {
					const hook =
						/** @type {NormalModuleCompilationHooks***REMOVED*** */
						(hooks).readResource.for(scheme);
					return createFakeHook(
						/** @type {AsyncSeriesBailHook<[string, NormalModule], string | Buffer | null>***REMOVED*** */ ({
							tap: (options, fn) =>
								hook.tap(options, loaderContext =>
									fn(
										loaderContext.resource,
										/** @type {NormalModule***REMOVED*** */ (loaderContext._module)
									)
								),
							tapAsync: (options, fn) =>
								hook.tapAsync(options, (loaderContext, callback) =>
									fn(
										loaderContext.resource,
										/** @type {NormalModule***REMOVED*** */ (loaderContext._module),
										callback
									)
								),
							tapPromise: (options, fn) =>
								hook.tapPromise(options, loaderContext =>
									fn(
										loaderContext.resource,
										/** @type {NormalModule***REMOVED*** */ (loaderContext._module)
									)
								)
						***REMOVED***)
					);
				***REMOVED***),
				readResource: new HookMap(
					() => new AsyncSeriesBailHook(["loaderContext"])
				),
				processResult: new SyncWaterfallHook(["result", "module"]),
				needBuild: new AsyncSeriesBailHook(["module", "context"])
			***REMOVED***;
			compilationHooksMap.set(
				compilation,
				/** @type {NormalModuleCompilationHooks***REMOVED*** */ (hooks)
			);
		***REMOVED***
		return /** @type {NormalModuleCompilationHooks***REMOVED*** */ (hooks);
	***REMOVED***

	/**
	 * @param {NormalModuleCreateData***REMOVED*** options options object
	 */
	constructor({
		layer,
		type,
		request,
		userRequest,
		rawRequest,
		loaders,
		resource,
		resourceResolveData,
		context,
		matchResource,
		parser,
		parserOptions,
		generator,
		generatorOptions,
		resolveOptions
	***REMOVED***) {
		super(type, context || getContext(resource), layer);

		// Info from Factory
		/** @type {string***REMOVED*** */
		this.request = request;
		/** @type {string***REMOVED*** */
		this.userRequest = userRequest;
		/** @type {string***REMOVED*** */
		this.rawRequest = rawRequest;
		/** @type {boolean***REMOVED*** */
		this.binary = /^(asset|webassembly)\b/.test(type);
		/** @type {undefined | Parser***REMOVED*** */
		this.parser = parser;
		/** @type {undefined | ParserOptions***REMOVED*** */
		this.parserOptions = parserOptions;
		/** @type {undefined | Generator***REMOVED*** */
		this.generator = generator;
		/** @type {undefined | GeneratorOptions***REMOVED*** */
		this.generatorOptions = generatorOptions;
		/** @type {string***REMOVED*** */
		this.resource = resource;
		this.resourceResolveData = resourceResolveData;
		/** @type {string | undefined***REMOVED*** */
		this.matchResource = matchResource;
		/** @type {LoaderItem[]***REMOVED*** */
		this.loaders = loaders;
		if (resolveOptions !== undefined) {
			// already declared in super class
			this.resolveOptions = resolveOptions;
		***REMOVED***

		// Info from Build
		/** @type {WebpackError | null***REMOVED*** */
		this.error = null;
		/**
		 * @private
		 * @type {Source | null***REMOVED***
		 */
		this._source = null;
		/**
		 * @private
		 * @type {Map<string | undefined, number> | undefined***REMOVED***
		 */
		this._sourceSizes = undefined;
		/**
		 * @private
		 * @type {undefined | SourceTypes***REMOVED***
		 */
		this._sourceTypes = undefined;

		// Cache
		this._lastSuccessfulBuildMeta = {***REMOVED***;
		this._forceBuild = true;
		this._isEvaluatingSideEffects = false;
		/** @type {WeakSet<ModuleGraph> | undefined***REMOVED*** */
		this._addedSideEffectsBailout = undefined;
		/** @type {CodeGeneratorData***REMOVED*** */
		this._codeGeneratorData = new Map();
	***REMOVED***

	/**
	 * @returns {string***REMOVED*** a unique identifier of the module
	 */
	identifier() {
		if (this.layer === null) {
			if (this.type === JAVASCRIPT_MODULE_TYPE_AUTO) {
				return this.request;
			***REMOVED***
			return `${this.type***REMOVED***|${this.request***REMOVED***`;
		***REMOVED***
		return `${this.type***REMOVED***|${this.request***REMOVED***|${this.layer***REMOVED***`;
	***REMOVED***

	/**
	 * @param {RequestShortener***REMOVED*** requestShortener the request shortener
	 * @returns {string***REMOVED*** a user readable identifier of the module
	 */
	readableIdentifier(requestShortener) {
		return /** @type {string***REMOVED*** */ (requestShortener.shorten(this.userRequest));
	***REMOVED***

	/**
	 * @param {LibIdentOptions***REMOVED*** options options
	 * @returns {string | null***REMOVED*** an identifier for library inclusion
	 */
	libIdent(options) {
		let ident = contextify(
			options.context,
			this.userRequest,
			options.associatedObjectForCache
		);
		if (this.layer) ident = `(${this.layer***REMOVED***)/${ident***REMOVED***`;
		return ident;
	***REMOVED***

	/**
	 * @returns {string | null***REMOVED*** absolute path which should be used for condition matching (usually the resource path)
	 */
	nameForCondition() {
		const resource = this.matchResource || this.resource;
		const idx = resource.indexOf("?");
		if (idx >= 0) return resource.slice(0, idx);
		return resource;
	***REMOVED***

	/**
	 * Assuming this module is in the cache. Update the (cached) module with
	 * the fresh module from the factory. Usually updates internal references
	 * and properties.
	 * @param {Module***REMOVED*** module fresh module
	 * @returns {void***REMOVED***
	 */
	updateCacheModule(module) {
		super.updateCacheModule(module);
		const m = /** @type {NormalModule***REMOVED*** */ (module);
		this.binary = m.binary;
		this.request = m.request;
		this.userRequest = m.userRequest;
		this.rawRequest = m.rawRequest;
		this.parser = m.parser;
		this.parserOptions = m.parserOptions;
		this.generator = m.generator;
		this.generatorOptions = m.generatorOptions;
		this.resource = m.resource;
		this.resourceResolveData = m.resourceResolveData;
		this.context = m.context;
		this.matchResource = m.matchResource;
		this.loaders = m.loaders;
	***REMOVED***

	/**
	 * Assuming this module is in the cache. Remove internal references to allow freeing some memory.
	 */
	cleanupForCache() {
		// Make sure to cache types and sizes before cleanup when this module has been built
		// They are accessed by the stats and we don't want them to crash after cleanup
		// TODO reconsider this for webpack 6
		if (this.buildInfo) {
			if (this._sourceTypes === undefined) this.getSourceTypes();
			for (const type of /** @type {SourceTypes***REMOVED*** */ (this._sourceTypes)) {
				this.size(type);
			***REMOVED***
		***REMOVED***
		super.cleanupForCache();
		this.parser = undefined;
		this.parserOptions = undefined;
		this.generator = undefined;
		this.generatorOptions = undefined;
	***REMOVED***

	/**
	 * Module should be unsafe cached. Get data that's needed for that.
	 * This data will be passed to restoreFromUnsafeCache later.
	 * @returns {UnsafeCacheData***REMOVED*** cached data
	 */
	getUnsafeCacheData() {
		const data = super.getUnsafeCacheData();
		data.parserOptions = this.parserOptions;
		data.generatorOptions = this.generatorOptions;
		return data;
	***REMOVED***

	/**
	 * restore unsafe cache data
	 * @param {UnsafeCacheData***REMOVED*** unsafeCacheData data from getUnsafeCacheData
	 * @param {NormalModuleFactory***REMOVED*** normalModuleFactory the normal module factory handling the unsafe caching
	 */
	restoreFromUnsafeCache(unsafeCacheData, normalModuleFactory) {
		this._restoreFromUnsafeCache(unsafeCacheData, normalModuleFactory);
	***REMOVED***

	/**
	 * restore unsafe cache data
	 * @param {UnsafeCacheData***REMOVED*** unsafeCacheData data from getUnsafeCacheData
	 * @param {NormalModuleFactory***REMOVED*** normalModuleFactory the normal module factory handling the unsafe caching
	 */
	_restoreFromUnsafeCache(unsafeCacheData, normalModuleFactory) {
		super._restoreFromUnsafeCache(unsafeCacheData, normalModuleFactory);
		this.parserOptions = unsafeCacheData.parserOptions;
		this.parser = normalModuleFactory.getParser(this.type, this.parserOptions);
		this.generatorOptions = unsafeCacheData.generatorOptions;
		this.generator = normalModuleFactory.getGenerator(
			this.type,
			this.generatorOptions
		);
		// we assume the generator behaves identically and keep cached sourceTypes/Sizes
	***REMOVED***

	/**
	 * @param {string***REMOVED*** context the compilation context
	 * @param {string***REMOVED*** name the asset name
	 * @param {string | Buffer***REMOVED*** content the content
	 * @param {(string | SourceMap)=***REMOVED*** sourceMap an optional source map
	 * @param {AssociatedObjectForCache=***REMOVED*** associatedObjectForCache object for caching
	 * @returns {Source***REMOVED*** the created source
	 */
	createSourceForAsset(
		context,
		name,
		content,
		sourceMap,
		associatedObjectForCache
	) {
		if (sourceMap) {
			if (
				typeof sourceMap === "string" &&
				(this.useSourceMap || this.useSimpleSourceMap)
			) {
				return new OriginalSource(
					content,
					contextifySourceUrl(context, sourceMap, associatedObjectForCache)
				);
			***REMOVED***

			if (this.useSourceMap) {
				return new SourceMapSource(
					content,
					name,
					contextifySourceMap(
						context,
						/** @type {SourceMap***REMOVED*** */ (sourceMap),
						associatedObjectForCache
					)
				);
			***REMOVED***
		***REMOVED***

		return new RawSource(content);
	***REMOVED***

	/**
	 * @private
	 * @template T
	 * @param {ResolverWithOptions***REMOVED*** resolver a resolver
	 * @param {WebpackOptions***REMOVED*** options webpack options
	 * @param {Compilation***REMOVED*** compilation the compilation
	 * @param {InputFileSystem***REMOVED*** fs file system from reading
	 * @param {NormalModuleCompilationHooks***REMOVED*** hooks the hooks
	 * @returns {import("../declarations/LoaderContext").NormalModuleLoaderContext<T>***REMOVED*** loader context
	 */
	_createLoaderContext(resolver, options, compilation, fs, hooks) {
		const { requestShortener ***REMOVED*** = compilation.runtimeTemplate;
		const getCurrentLoaderName = () => {
			const currentLoader = this.getCurrentLoader(loaderContext);
			if (!currentLoader) return "(not in loader scope)";
			return requestShortener.shorten(currentLoader.loader);
		***REMOVED***;
		/**
		 * @returns {ResolveContext***REMOVED*** resolve context
		 */
		const getResolveContext = () => ({
			fileDependencies: {
				add: d => /** @type {TODO***REMOVED*** */ (loaderContext).addDependency(d)
			***REMOVED***,
			contextDependencies: {
				add: d => /** @type {TODO***REMOVED*** */ (loaderContext).addContextDependency(d)
			***REMOVED***,
			missingDependencies: {
				add: d => /** @type {TODO***REMOVED*** */ (loaderContext).addMissingDependency(d)
			***REMOVED***
		***REMOVED***);
		const getAbsolutify = memoize(() =>
			absolutify.bindCache(compilation.compiler.root)
		);
		const getAbsolutifyInContext = memoize(() =>
			absolutify.bindContextCache(
				/** @type {string***REMOVED*** */
				(this.context),
				compilation.compiler.root
			)
		);
		const getContextify = memoize(() =>
			contextify.bindCache(compilation.compiler.root)
		);
		const getContextifyInContext = memoize(() =>
			contextify.bindContextCache(
				/** @type {string***REMOVED*** */
				(this.context),
				compilation.compiler.root
			)
		);
		const utils = {
			/**
			 * @param {string***REMOVED*** context context
			 * @param {string***REMOVED*** request request
			 * @returns {string***REMOVED*** result
			 */
			absolutify: (context, request) =>
				context === this.context
					? getAbsolutifyInContext()(request)
					: getAbsolutify()(context, request),
			/**
			 * @param {string***REMOVED*** context context
			 * @param {string***REMOVED*** request request
			 * @returns {string***REMOVED*** result
			 */
			contextify: (context, request) =>
				context === this.context
					? getContextifyInContext()(request)
					: getContextify()(context, request),
			/**
			 * @param {HashFunction=***REMOVED*** type type
			 * @returns {Hash***REMOVED*** hash
			 */
			createHash: type =>
				createHash(
					type ||
						/** @type {HashFunction***REMOVED*** */
						(compilation.outputOptions.hashFunction)
				)
		***REMOVED***;
		/** @type {import("../declarations/LoaderContext").NormalModuleLoaderContext<T>***REMOVED*** */
		const loaderContext = {
			version: 2,
			/**
			 * @param {import("../declarations/LoaderContext").Schema=***REMOVED*** schema schema
			 * @returns {T***REMOVED*** options
			 */
			getOptions: schema => {
				const loader = this.getCurrentLoader(loaderContext);

				let { options ***REMOVED*** = /** @type {LoaderItem***REMOVED*** */ (loader);

				if (typeof options === "string") {
					if (options.startsWith("{") && options.endsWith("***REMOVED***")) {
						try {
							options = parseJson(options);
						***REMOVED*** catch (err) {
							throw new Error(
								`Cannot parse string options: ${/** @type {Error***REMOVED*** */ (err).message***REMOVED***`
							);
						***REMOVED***
					***REMOVED*** else {
						options = querystring.parse(options, "&", "=", {
							maxKeys: 0
						***REMOVED***);
					***REMOVED***
				***REMOVED***

				if (options === null || options === undefined) {
					options = {***REMOVED***;
				***REMOVED***

				if (schema) {
					let name = "Loader";
					let baseDataPath = "options";
					let match;
					if (schema.title && (match = /^(.+) (.+)$/.exec(schema.title))) {
						[, name, baseDataPath] = match;
					***REMOVED***
					getValidate()(schema, /** @type {EXPECTED_OBJECT***REMOVED*** */ (options), {
						name,
						baseDataPath
					***REMOVED***);
				***REMOVED***

				return /** @type {T***REMOVED*** */ (options);
			***REMOVED***,
			emitWarning: warning => {
				if (!(warning instanceof Error)) {
					warning = new NonErrorEmittedError(warning);
				***REMOVED***
				this.addWarning(
					new ModuleWarning(warning, {
						from: getCurrentLoaderName()
					***REMOVED***)
				);
			***REMOVED***,
			emitError: error => {
				if (!(error instanceof Error)) {
					error = new NonErrorEmittedError(error);
				***REMOVED***
				this.addError(
					new ModuleError(error, {
						from: getCurrentLoaderName()
					***REMOVED***)
				);
			***REMOVED***,
			getLogger: name => {
				const currentLoader = this.getCurrentLoader(loaderContext);
				return compilation.getLogger(() =>
					[currentLoader && currentLoader.loader, name, this.identifier()]
						.filter(Boolean)
						.join("|")
				);
			***REMOVED***,
			resolve(context, request, callback) {
				resolver.resolve({***REMOVED***, context, request, getResolveContext(), callback);
			***REMOVED***,
			getResolve(options) {
				const child = options ? resolver.withOptions(options) : resolver;
				return /** @type {ReturnType<import("../declarations/LoaderContext").NormalModuleLoaderContext<T>["getResolve"]>***REMOVED*** */ (
					(context, request, callback) => {
						if (callback) {
							child.resolve(
								{***REMOVED***,
								context,
								request,
								getResolveContext(),
								callback
							);
						***REMOVED*** else {
							return new Promise((resolve, reject) => {
								child.resolve(
									{***REMOVED***,
									context,
									request,
									getResolveContext(),
									(err, result) => {
										if (err) reject(err);
										else resolve(result);
									***REMOVED***
								);
							***REMOVED***);
						***REMOVED***
					***REMOVED***
				);
			***REMOVED***,
			emitFile: (name, content, sourceMap, assetInfo) => {
				const buildInfo = /** @type {BuildInfo***REMOVED*** */ (this.buildInfo);

				if (!buildInfo.assets) {
					buildInfo.assets = Object.create(null);
					buildInfo.assetsInfo = new Map();
				***REMOVED***

				const assets =
					/** @type {NonNullable<KnownBuildInfo["assets"]>***REMOVED*** */
					(buildInfo.assets);
				const assetsInfo =
					/** @type {NonNullable<KnownBuildInfo["assetsInfo"]>***REMOVED*** */
					(buildInfo.assetsInfo);

				assets[name] = this.createSourceForAsset(
					/** @type {string***REMOVED*** */ (options.context),
					name,
					content,
					sourceMap,
					compilation.compiler.root
				);
				assetsInfo.set(name, assetInfo);
			***REMOVED***,
			addBuildDependency: dep => {
				const buildInfo = /** @type {BuildInfo***REMOVED*** */ (this.buildInfo);

				if (buildInfo.buildDependencies === undefined) {
					buildInfo.buildDependencies = new LazySet();
				***REMOVED***
				buildInfo.buildDependencies.add(dep);
			***REMOVED***,
			utils,
			rootContext: /** @type {string***REMOVED*** */ (options.context),
			webpack: true,
			sourceMap: Boolean(this.useSourceMap),
			mode: options.mode || "production",
			hashFunction: /** @type {string***REMOVED*** */ (options.output.hashFunction),
			hashDigest: /** @type {string***REMOVED*** */ (options.output.hashDigest),
			hashDigestLength: /** @type {number***REMOVED*** */ (options.output.hashDigestLength),
			hashSalt: /** @type {string***REMOVED*** */ (options.output.hashSalt),
			_module: this,
			_compilation: compilation,
			_compiler: compilation.compiler,
			fs
		***REMOVED***;

		Object.assign(loaderContext, options.loader);

		hooks.loader.call(
			/** @type {LoaderContext<EXPECTED_ANY>***REMOVED*** */
			(loaderContext),
			this
		);

		return loaderContext;
	***REMOVED***

	// TODO remove `loaderContext` in webpack@6
	/**
	 * @param {TODO***REMOVED*** loaderContext loader context
	 * @param {number***REMOVED*** index index
	 * @returns {LoaderItem | null***REMOVED*** loader
	 */
	getCurrentLoader(loaderContext, index = loaderContext.loaderIndex) {
		if (
			this.loaders &&
			this.loaders.length &&
			index < this.loaders.length &&
			index >= 0 &&
			this.loaders[index]
		) {
			return this.loaders[index];
		***REMOVED***
		return null;
	***REMOVED***

	/**
	 * @param {string***REMOVED*** context the compilation context
	 * @param {string | Buffer***REMOVED*** content the content
	 * @param {(string | SourceMapSource | null)=***REMOVED*** sourceMap an optional source map
	 * @param {AssociatedObjectForCache=***REMOVED*** associatedObjectForCache object for caching
	 * @returns {Source***REMOVED*** the created source
	 */
	createSource(context, content, sourceMap, associatedObjectForCache) {
		if (Buffer.isBuffer(content)) {
			return new RawSource(content);
		***REMOVED***

		// if there is no identifier return raw source
		if (!this.identifier) {
			return new RawSource(content);
		***REMOVED***

		// from here on we assume we have an identifier
		const identifier = this.identifier();

		if (this.useSourceMap && sourceMap) {
			return new SourceMapSource(
				content,
				contextifySourceUrl(context, identifier, associatedObjectForCache),
				contextifySourceMap(
					context,
					/** @type {TODO***REMOVED*** */ (sourceMap),
					associatedObjectForCache
				)
			);
		***REMOVED***

		if (this.useSourceMap || this.useSimpleSourceMap) {
			return new OriginalSource(
				content,
				contextifySourceUrl(context, identifier, associatedObjectForCache)
			);
		***REMOVED***

		return new RawSource(content);
	***REMOVED***

	/**
	 * @param {WebpackOptions***REMOVED*** options webpack options
	 * @param {Compilation***REMOVED*** compilation the compilation
	 * @param {ResolverWithOptions***REMOVED*** resolver the resolver
	 * @param {InputFileSystem***REMOVED*** fs the file system
	 * @param {NormalModuleCompilationHooks***REMOVED*** hooks the hooks
	 * @param {BuildCallback***REMOVED*** callback callback function
	 * @returns {void***REMOVED***
	 */
	_doBuild(options, compilation, resolver, fs, hooks, callback) {
		const loaderContext = this._createLoaderContext(
			resolver,
			options,
			compilation,
			fs,
			hooks
		);

		/**
		 * @param {Error | null***REMOVED*** err err
		 * @param {(Result | null)=***REMOVED*** _result result
		 * @returns {void***REMOVED***
		 */
		const processResult = (err, _result) => {
			if (err) {
				if (!(err instanceof Error)) {
					err = new NonErrorEmittedError(err);
				***REMOVED***
				const currentLoader = this.getCurrentLoader(loaderContext);
				const error = new ModuleBuildError(err, {
					from:
						currentLoader &&
						compilation.runtimeTemplate.requestShortener.shorten(
							currentLoader.loader
						)
				***REMOVED***);
				return callback(error);
			***REMOVED***
			const result = hooks.processResult.call(
				/** @type {Result***REMOVED*** */ (_result),
				this
			);
			const source = result[0];
			const sourceMap = result.length >= 1 ? result[1] : null;
			const extraInfo = result.length >= 2 ? result[2] : null;

			if (!Buffer.isBuffer(source) && typeof source !== "string") {
				const currentLoader = this.getCurrentLoader(loaderContext, 0);
				const err = new Error(
					`Final loader (${
						currentLoader
							? compilation.runtimeTemplate.requestShortener.shorten(
									currentLoader.loader
								)
							: "unknown"
					***REMOVED***) didn't return a Buffer or String`
				);
				const error = new ModuleBuildError(err);
				return callback(error);
			***REMOVED***

			const isBinaryModule =
				this.generatorOptions && this.generatorOptions.binary !== undefined
					? this.generatorOptions.binary
					: this.binary;

			this._source = this.createSource(
				/** @type {string***REMOVED*** */ (options.context),
				isBinaryModule ? asBuffer(source) : asString(source),
				sourceMap,
				compilation.compiler.root
			);
			if (this._sourceSizes !== undefined) this._sourceSizes.clear();
			this._ast =
				typeof extraInfo === "object" &&
				extraInfo !== null &&
				extraInfo.webpackAST !== undefined
					? extraInfo.webpackAST
					: null;
			return callback();
		***REMOVED***;

		const buildInfo = /** @type {BuildInfo***REMOVED*** */ (this.buildInfo);

		buildInfo.fileDependencies = new LazySet();
		buildInfo.contextDependencies = new LazySet();
		buildInfo.missingDependencies = new LazySet();
		buildInfo.cacheable = true;

		try {
			hooks.beforeLoaders.call(
				this.loaders,
				this,
				/** @type {LoaderContext<EXPECTED_ANY>***REMOVED*** */
				(loaderContext)
			);
		***REMOVED*** catch (err) {
			processResult(/** @type {Error***REMOVED*** */ (err));
			return;
		***REMOVED***

		if (this.loaders.length > 0) {
			/** @type {BuildInfo***REMOVED*** */
			(this.buildInfo).buildDependencies = new LazySet();
		***REMOVED***

		runLoaders(
			{
				resource: this.resource,
				loaders: this.loaders,
				context: loaderContext,
				/**
				 * @param {LoaderContext<TODO>***REMOVED*** loaderContext the loader context
				 * @param {string***REMOVED*** resourcePath the resource Path
				 * @param {(err: Error | null, result?: string | Buffer) => void***REMOVED*** callback callback
				 */
				processResource: (loaderContext, resourcePath, callback) => {
					const resource = loaderContext.resource;
					const scheme = getScheme(resource);
					hooks.readResource
						.for(scheme)
						.callAsync(loaderContext, (err, result) => {
							if (err) return callback(err);
							if (typeof result !== "string" && !result) {
								return callback(
									new UnhandledSchemeError(
										/** @type {string***REMOVED*** */
										(scheme),
										resource
									)
								);
							***REMOVED***
							return callback(null, result);
						***REMOVED***);
				***REMOVED***
			***REMOVED***,
			(err, result) => {
				// Cleanup loaderContext to avoid leaking memory in ICs
				loaderContext._compilation =
					loaderContext._compiler =
					loaderContext._module =
					loaderContext.fs =
						/** @type {EXPECTED_ANY***REMOVED*** */
						(undefined);

				if (!result) {
					/** @type {BuildInfo***REMOVED*** */
					(this.buildInfo).cacheable = false;
					return processResult(
						err || new Error("No result from loader-runner processing"),
						null
					);
				***REMOVED***

				const buildInfo = /** @type {BuildInfo***REMOVED*** */ (this.buildInfo);

				const fileDependencies =
					/** @type {NonNullable<KnownBuildInfo["fileDependencies"]>***REMOVED*** */
					(buildInfo.fileDependencies);
				const contextDependencies =
					/** @type {NonNullable<KnownBuildInfo["contextDependencies"]>***REMOVED*** */
					(buildInfo.contextDependencies);
				const missingDependencies =
					/** @type {NonNullable<KnownBuildInfo["missingDependencies"]>***REMOVED*** */
					(buildInfo.missingDependencies);

				fileDependencies.addAll(result.fileDependencies);
				contextDependencies.addAll(result.contextDependencies);
				missingDependencies.addAll(result.missingDependencies);
				for (const loader of this.loaders) {
					const buildDependencies =
						/** @type {NonNullable<KnownBuildInfo["buildDependencies"]>***REMOVED*** */
						(buildInfo.buildDependencies);

					buildDependencies.add(loader.loader);
				***REMOVED***
				buildInfo.cacheable = buildInfo.cacheable && result.cacheable;
				processResult(err, result.result);
			***REMOVED***
		);
	***REMOVED***

	/**
	 * @param {WebpackError***REMOVED*** error the error
	 * @returns {void***REMOVED***
	 */
	markModuleAsErrored(error) {
		// Restore build meta from successful build to keep importing state
		this.buildMeta = { ...this._lastSuccessfulBuildMeta ***REMOVED***;
		this.error = error;
		this.addError(error);
	***REMOVED***

	/**
	 * @param {Exclude<NoParse, EXPECTED_ANY[]>***REMOVED*** rule rule
	 * @param {string***REMOVED*** content content
	 * @returns {boolean***REMOVED*** result
	 */
	applyNoParseRule(rule, content) {
		// must start with "rule" if rule is a string
		if (typeof rule === "string") {
			return content.startsWith(rule);
		***REMOVED***

		if (typeof rule === "function") {
			return rule(content);
		***REMOVED***
		// we assume rule is a regexp
		return rule.test(content);
	***REMOVED***

	/**
	 * @param {undefined | NoParse***REMOVED*** noParseRule no parse rule
	 * @param {string***REMOVED*** request request
	 * @returns {boolean***REMOVED*** check if module should not be parsed, returns "true" if the module should !not! be parsed, returns "false" if the module !must! be parsed
	 */
	shouldPreventParsing(noParseRule, request) {
		// if no noParseRule exists, return false
		// the module !must! be parsed.
		if (!noParseRule) {
			return false;
		***REMOVED***

		// we only have one rule to check
		if (!Array.isArray(noParseRule)) {
			// returns "true" if the module is !not! to be parsed
			return this.applyNoParseRule(noParseRule, request);
		***REMOVED***

		for (let i = 0; i < noParseRule.length; i++) {
			const rule = noParseRule[i];
			// early exit on first truthy match
			// this module is !not! to be parsed
			if (this.applyNoParseRule(rule, request)) {
				return true;
			***REMOVED***
		***REMOVED***
		// no match found, so this module !should! be parsed
		return false;
	***REMOVED***

	/**
	 * @param {Compilation***REMOVED*** compilation compilation
	 * @private
	 */
	_initBuildHash(compilation) {
		const hash = createHash(
			/** @type {HashFunction***REMOVED*** */
			(compilation.outputOptions.hashFunction)
		);
		if (this._source) {
			hash.update("source");
			this._source.updateHash(hash);
		***REMOVED***
		hash.update("meta");
		hash.update(JSON.stringify(this.buildMeta));
		/** @type {BuildInfo***REMOVED*** */
		(this.buildInfo).hash = /** @type {string***REMOVED*** */ (hash.digest("hex"));
	***REMOVED***

	/**
	 * @param {WebpackOptions***REMOVED*** options webpack options
	 * @param {Compilation***REMOVED*** compilation the compilation
	 * @param {ResolverWithOptions***REMOVED*** resolver the resolver
	 * @param {InputFileSystem***REMOVED*** fs the file system
	 * @param {BuildCallback***REMOVED*** callback callback function
	 * @returns {void***REMOVED***
	 */
	build(options, compilation, resolver, fs, callback) {
		this._forceBuild = false;
		this._source = null;
		if (this._sourceSizes !== undefined) this._sourceSizes.clear();
		this._sourceTypes = undefined;
		this._ast = null;
		this.error = null;
		this.clearWarningsAndErrors();
		this.clearDependenciesAndBlocks();
		this.buildMeta = {***REMOVED***;
		this.buildInfo = {
			cacheable: false,
			parsed: true,
			fileDependencies: undefined,
			contextDependencies: undefined,
			missingDependencies: undefined,
			buildDependencies: undefined,
			valueDependencies: undefined,
			hash: undefined,
			assets: undefined,
			assetsInfo: undefined
		***REMOVED***;

		const startTime = compilation.compiler.fsStartTime || Date.now();

		const hooks = NormalModule.getCompilationHooks(compilation);

		return this._doBuild(options, compilation, resolver, fs, hooks, err => {
			// if we have an error mark module as failed and exit
			if (err) {
				this.markModuleAsErrored(err);
				this._initBuildHash(compilation);
				return callback();
			***REMOVED***

			/**
			 * @param {Error***REMOVED*** e error
			 * @returns {void***REMOVED***
			 */
			const handleParseError = e => {
				const source = /** @type {Source***REMOVED*** */ (this._source).source();
				const loaders = this.loaders.map(item =>
					contextify(
						/** @type {string***REMOVED*** */ (options.context),
						item.loader,
						compilation.compiler.root
					)
				);
				const error = new ModuleParseError(source, e, loaders, this.type);
				this.markModuleAsErrored(error);
				this._initBuildHash(compilation);
				return callback();
			***REMOVED***;

			const handleParseResult = () => {
				this.dependencies.sort(
					concatComparators(
						compareSelect(a => a.loc, compareLocations),
						keepOriginalOrder(this.dependencies)
					)
				);
				this._initBuildHash(compilation);
				this._lastSuccessfulBuildMeta =
					/** @type {BuildMeta***REMOVED*** */
					(this.buildMeta);
				return handleBuildDone();
			***REMOVED***;

			const handleBuildDone = () => {
				try {
					hooks.beforeSnapshot.call(this);
				***REMOVED*** catch (err) {
					this.markModuleAsErrored(/** @type {WebpackError***REMOVED*** */ (err));
					return callback();
				***REMOVED***

				const snapshotOptions = compilation.options.snapshot.module;
				const { cacheable ***REMOVED*** = /** @type {BuildInfo***REMOVED*** */ (this.buildInfo);
				if (!cacheable || !snapshotOptions) {
					return callback();
				***REMOVED***
				// add warning for all non-absolute paths in fileDependencies, etc
				// This makes it easier to find problems with watching and/or caching
				/** @type {undefined | Set<string>***REMOVED*** */
				let nonAbsoluteDependencies;
				/**
				 * @param {LazySet<string>***REMOVED*** deps deps
				 */
				const checkDependencies = deps => {
					for (const dep of deps) {
						if (!ABSOLUTE_PATH_REGEX.test(dep)) {
							if (nonAbsoluteDependencies === undefined)
								nonAbsoluteDependencies = new Set();
							nonAbsoluteDependencies.add(dep);
							deps.delete(dep);
							try {
								const depWithoutGlob = dep.replace(/[\\/]?\*.*$/, "");
								const absolute = join(
									compilation.fileSystemInfo.fs,
									/** @type {string***REMOVED*** */
									(this.context),
									depWithoutGlob
								);
								if (absolute !== dep && ABSOLUTE_PATH_REGEX.test(absolute)) {
									(depWithoutGlob !== dep
										? /** @type {NonNullable<KnownBuildInfo["contextDependencies"]>***REMOVED*** */
											(
												/** @type {BuildInfo***REMOVED*** */
												(this.buildInfo).contextDependencies
											)
										: deps
									).add(absolute);
								***REMOVED***
							***REMOVED*** catch (_err) {
								// ignore
							***REMOVED***
						***REMOVED***
					***REMOVED***
				***REMOVED***;
				const buildInfo = /** @type {BuildInfo***REMOVED*** */ (this.buildInfo);
				const fileDependencies =
					/** @type {NonNullable<KnownBuildInfo["fileDependencies"]>***REMOVED*** */
					(buildInfo.fileDependencies);
				const contextDependencies =
					/** @type {NonNullable<KnownBuildInfo["contextDependencies"]>***REMOVED*** */
					(buildInfo.contextDependencies);
				const missingDependencies =
					/** @type {NonNullable<KnownBuildInfo["missingDependencies"]>***REMOVED*** */
					(buildInfo.missingDependencies);
				checkDependencies(fileDependencies);
				checkDependencies(missingDependencies);
				checkDependencies(contextDependencies);
				if (nonAbsoluteDependencies !== undefined) {
					const InvalidDependenciesModuleWarning =
						getInvalidDependenciesModuleWarning();
					this.addWarning(
						new InvalidDependenciesModuleWarning(this, nonAbsoluteDependencies)
					);
				***REMOVED***
				// convert file/context/missingDependencies into filesystem snapshot
				compilation.fileSystemInfo.createSnapshot(
					startTime,
					fileDependencies,
					contextDependencies,
					missingDependencies,
					snapshotOptions,
					(err, snapshot) => {
						if (err) {
							this.markModuleAsErrored(err);
							return;
						***REMOVED***
						buildInfo.fileDependencies = undefined;
						buildInfo.contextDependencies = undefined;
						buildInfo.missingDependencies = undefined;
						buildInfo.snapshot = snapshot;
						return callback();
					***REMOVED***
				);
			***REMOVED***;

			try {
				hooks.beforeParse.call(this);
			***REMOVED*** catch (err) {
				this.markModuleAsErrored(/** @type {WebpackError***REMOVED*** */ (err));
				this._initBuildHash(compilation);
				return callback();
			***REMOVED***

			// check if this module should !not! be parsed.
			// if so, exit here;
			const noParseRule = options.module && options.module.noParse;
			if (this.shouldPreventParsing(noParseRule, this.request)) {
				// We assume that we need module and exports
				/** @type {BuildInfo***REMOVED*** */
				(this.buildInfo).parsed = false;
				this._initBuildHash(compilation);
				return handleBuildDone();
			***REMOVED***

			try {
				const source = /** @type {Source***REMOVED*** */ (this._source).source();
				/** @type {Parser***REMOVED*** */
				(this.parser).parse(this._ast || source, {
					source,
					current: this,
					module: this,
					compilation,
					options
				***REMOVED***);
			***REMOVED*** catch (parseErr) {
				handleParseError(/** @type {Error***REMOVED*** */ (parseErr));
				return;
			***REMOVED***
			handleParseResult();
		***REMOVED***);
	***REMOVED***

	/**
	 * @param {ConcatenationBailoutReasonContext***REMOVED*** context context
	 * @returns {string | undefined***REMOVED*** reason why this module can't be concatenated, undefined when it can be concatenated
	 */
	getConcatenationBailoutReason(context) {
		return /** @type {Generator***REMOVED*** */ (
			this.generator
		).getConcatenationBailoutReason(this, context);
	***REMOVED***

	/**
	 * @param {ModuleGraph***REMOVED*** moduleGraph the module graph
	 * @returns {ConnectionState***REMOVED*** how this module should be connected to referencing modules when consumed for side-effects only
	 */
	getSideEffectsConnectionState(moduleGraph) {
		if (this.factoryMeta !== undefined) {
			if (this.factoryMeta.sideEffectFree) return false;
			if (this.factoryMeta.sideEffectFree === false) return true;
		***REMOVED***
		if (this.buildMeta !== undefined && this.buildMeta.sideEffectFree) {
			if (this._isEvaluatingSideEffects)
				return ModuleGraphConnection.CIRCULAR_CONNECTION;
			this._isEvaluatingSideEffects = true;
			/** @type {ConnectionState***REMOVED*** */
			let current = false;
			for (const dep of this.dependencies) {
				const state = dep.getModuleEvaluationSideEffectsState(moduleGraph);
				if (state === true) {
					if (
						this._addedSideEffectsBailout === undefined
							? ((this._addedSideEffectsBailout = new WeakSet()), true)
							: !this._addedSideEffectsBailout.has(moduleGraph)
					) {
						this._addedSideEffectsBailout.add(moduleGraph);
						moduleGraph
							.getOptimizationBailout(this)
							.push(
								() =>
									`Dependency (${
										dep.type
									***REMOVED***) with side effects at ${formatLocation(dep.loc)***REMOVED***`
							);
					***REMOVED***
					this._isEvaluatingSideEffects = false;
					return true;
				***REMOVED*** else if (state !== ModuleGraphConnection.CIRCULAR_CONNECTION) {
					current = ModuleGraphConnection.addConnectionStates(current, state);
				***REMOVED***
			***REMOVED***
			this._isEvaluatingSideEffects = false;
			// When caching is implemented here, make sure to not cache when
			// at least one circular connection was in the loop above
			return current;
		***REMOVED***
		return true;
	***REMOVED***

	/**
	 * @returns {SourceTypes***REMOVED*** types available (do not mutate)
	 */
	getSourceTypes() {
		if (this._sourceTypes === undefined) {
			this._sourceTypes = /** @type {Generator***REMOVED*** */ (this.generator).getTypes(
				this
			);
		***REMOVED***
		return this._sourceTypes;
	***REMOVED***

	/**
	 * @param {CodeGenerationContext***REMOVED*** context context for code generation
	 * @returns {CodeGenerationResult***REMOVED*** result
	 */
	codeGeneration({
		dependencyTemplates,
		runtimeTemplate,
		moduleGraph,
		chunkGraph,
		runtime,
		concatenationScope,
		codeGenerationResults,
		sourceTypes
	***REMOVED***) {
		/** @type {Set<string>***REMOVED*** */
		const runtimeRequirements = new Set();

		const { parsed ***REMOVED*** = /** @type {BuildInfo***REMOVED*** */ (this.buildInfo);

		if (!parsed) {
			runtimeRequirements.add(RuntimeGlobals.module);
			runtimeRequirements.add(RuntimeGlobals.exports);
			runtimeRequirements.add(RuntimeGlobals.thisAsExports);
		***REMOVED***

		/**
		 * @type {() => CodeGeneratorData***REMOVED***
		 */
		const getData = () => this._codeGeneratorData;

		const sources = new Map();
		for (const type of sourceTypes || chunkGraph.getModuleSourceTypes(this)) {
			// TODO webpack@6 make generateError required
			const generator =
				/** @type {Generator & { generateError?: GenerateErrorFn ***REMOVED******REMOVED*** */
				(this.generator);
			const source = this.error
				? generator.generateError
					? generator.generateError(this.error, this, {
							dependencyTemplates,
							runtimeTemplate,
							moduleGraph,
							chunkGraph,
							runtimeRequirements,
							runtime,
							concatenationScope,
							codeGenerationResults,
							getData,
							type
						***REMOVED***)
					: new RawSource(
							`throw new Error(${JSON.stringify(this.error.message)***REMOVED***);`
						)
				: generator.generate(this, {
						dependencyTemplates,
						runtimeTemplate,
						moduleGraph,
						chunkGraph,
						runtimeRequirements,
						runtime,
						concatenationScope,
						codeGenerationResults,
						getData,
						type
					***REMOVED***);

			if (source) {
				sources.set(type, new CachedSource(source));
			***REMOVED***
		***REMOVED***

		/** @type {CodeGenerationResult***REMOVED*** */
		const resultEntry = {
			sources,
			runtimeRequirements,
			data: this._codeGeneratorData
		***REMOVED***;
		return resultEntry;
	***REMOVED***

	/**
	 * @returns {Source | null***REMOVED*** the original source for the module before webpack transformation
	 */
	originalSource() {
		return this._source;
	***REMOVED***

	/**
	 * @returns {void***REMOVED***
	 */
	invalidateBuild() {
		this._forceBuild = true;
	***REMOVED***

	/**
	 * @param {NeedBuildContext***REMOVED*** context context info
	 * @param {NeedBuildCallback***REMOVED*** callback callback function, returns true, if the module needs a rebuild
	 * @returns {void***REMOVED***
	 */
	needBuild(context, callback) {
		const { fileSystemInfo, compilation, valueCacheVersions ***REMOVED*** = context;
		// build if enforced
		if (this._forceBuild) return callback(null, true);

		// always try to build in case of an error
		if (this.error) return callback(null, true);

		const { cacheable, snapshot, valueDependencies ***REMOVED*** =
			/** @type {BuildInfo***REMOVED*** */ (this.buildInfo);

		// always build when module is not cacheable
		if (!cacheable) return callback(null, true);

		// build when there is no snapshot to check
		if (!snapshot) return callback(null, true);

		// build when valueDependencies have changed
		if (valueDependencies) {
			if (!valueCacheVersions) return callback(null, true);
			for (const [key, value] of valueDependencies) {
				if (value === undefined) return callback(null, true);
				const current = valueCacheVersions.get(key);
				if (
					value !== current &&
					(typeof value === "string" ||
						typeof current === "string" ||
						current === undefined ||
						!isSubset(value, current))
				) {
					return callback(null, true);
				***REMOVED***
			***REMOVED***
		***REMOVED***

		// check snapshot for validity
		fileSystemInfo.checkSnapshotValid(snapshot, (err, valid) => {
			if (err) return callback(err);
			if (!valid) return callback(null, true);
			const hooks = NormalModule.getCompilationHooks(compilation);
			hooks.needBuild.callAsync(this, context, (err, needBuild) => {
				if (err) {
					return callback(
						HookWebpackError.makeWebpackError(
							err,
							"NormalModule.getCompilationHooks().needBuild"
						)
					);
				***REMOVED***
				callback(null, Boolean(needBuild));
			***REMOVED***);
		***REMOVED***);
	***REMOVED***

	/**
	 * @param {string=***REMOVED*** type the source type for which the size should be estimated
	 * @returns {number***REMOVED*** the estimated size of the module (must be non-zero)
	 */
	size(type) {
		const cachedSize =
			this._sourceSizes === undefined ? undefined : this._sourceSizes.get(type);
		if (cachedSize !== undefined) {
			return cachedSize;
		***REMOVED***
		const size = Math.max(
			1,
			/** @type {Generator***REMOVED*** */ (this.generator).getSize(this, type)
		);
		if (this._sourceSizes === undefined) {
			this._sourceSizes = new Map();
		***REMOVED***
		this._sourceSizes.set(type, size);
		return size;
	***REMOVED***

	/**
	 * @param {LazySet<string>***REMOVED*** fileDependencies set where file dependencies are added to
	 * @param {LazySet<string>***REMOVED*** contextDependencies set where context dependencies are added to
	 * @param {LazySet<string>***REMOVED*** missingDependencies set where missing dependencies are added to
	 * @param {LazySet<string>***REMOVED*** buildDependencies set where build dependencies are added to
	 */
	addCacheDependencies(
		fileDependencies,
		contextDependencies,
		missingDependencies,
		buildDependencies
	) {
		const { snapshot, buildDependencies: buildDeps ***REMOVED*** =
			/** @type {BuildInfo***REMOVED*** */ (this.buildInfo);
		if (snapshot) {
			fileDependencies.addAll(snapshot.getFileIterable());
			contextDependencies.addAll(snapshot.getContextIterable());
			missingDependencies.addAll(snapshot.getMissingIterable());
		***REMOVED*** else {
			const {
				fileDependencies: fileDeps,
				contextDependencies: contextDeps,
				missingDependencies: missingDeps
			***REMOVED*** = /** @type {BuildInfo***REMOVED*** */ (this.buildInfo);
			if (fileDeps !== undefined) fileDependencies.addAll(fileDeps);
			if (contextDeps !== undefined) contextDependencies.addAll(contextDeps);
			if (missingDeps !== undefined) missingDependencies.addAll(missingDeps);
		***REMOVED***
		if (buildDeps !== undefined) {
			buildDependencies.addAll(buildDeps);
		***REMOVED***
	***REMOVED***

	/**
	 * @param {Hash***REMOVED*** hash the hash used to track dependencies
	 * @param {UpdateHashContext***REMOVED*** context context
	 * @returns {void***REMOVED***
	 */
	updateHash(hash, context) {
		const buildInfo = /** @type {BuildInfo***REMOVED*** */ (this.buildInfo);
		hash.update(
			/** @type {string***REMOVED*** */
			(buildInfo.hash)
		);
		/** @type {Generator***REMOVED*** */
		(this.generator).updateHash(hash, {
			module: this,
			...context
		***REMOVED***);
		super.updateHash(hash, context);
	***REMOVED***

	/**
	 * @param {ObjectSerializerContext***REMOVED*** context context
	 */
	serialize(context) {
		const { write ***REMOVED*** = context;
		// deserialize
		write(this._source);
		write(this.error);
		write(this._lastSuccessfulBuildMeta);
		write(this._forceBuild);
		write(this._codeGeneratorData);
		super.serialize(context);
	***REMOVED***

	/**
	 * @param {ObjectDeserializerContext***REMOVED*** context context
	 * @returns {TODO***REMOVED*** Module
	 */
	static deserialize(context) {
		const obj = new NormalModule({
			// will be deserialized by Module
			layer: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			type: "",
			// will be filled by updateCacheModule
			resource: "",
			context: "",
			request: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			userRequest: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			rawRequest: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			loaders: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			matchResource: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			parser: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			parserOptions: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			generator: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			generatorOptions: /** @type {EXPECTED_ANY***REMOVED*** */ (null),
			resolveOptions: /** @type {EXPECTED_ANY***REMOVED*** */ (null)
		***REMOVED***);
		obj.deserialize(context);
		return obj;
	***REMOVED***

	/**
	 * @param {ObjectDeserializerContext***REMOVED*** context context
	 */
	deserialize(context) {
		const { read ***REMOVED*** = context;
		this._source = read();
		this.error = read();
		this._lastSuccessfulBuildMeta = read();
		this._forceBuild = read();
		this._codeGeneratorData = read();
		super.deserialize(context);
	***REMOVED***
***REMOVED***

makeSerializable(NormalModule, "webpack/lib/NormalModule");

module.exports = NormalModule;
