// webpack/lib/Stats.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

/** @typedef {import("../declarations/WebpackOptions").StatsOptions***REMOVED*** StatsOptions */
/** @typedef {import("./Compilation")***REMOVED*** Compilation */
/** @typedef {import("./Compilation").NormalizedStatsOptions***REMOVED*** NormalizedStatsOptions */
/** @typedef {import("./stats/DefaultStatsFactoryPlugin").StatsCompilation***REMOVED*** StatsCompilation */

class Stats {
	/**
	 * @param {Compilation***REMOVED*** compilation webpack compilation
	 */
	constructor(compilation) {
		this.compilation = compilation;
	***REMOVED***

	get hash() {
		return this.compilation.hash;
	***REMOVED***

	get startTime() {
		return this.compilation.startTime;
	***REMOVED***

	get endTime() {
		return this.compilation.endTime;
	***REMOVED***

	/**
	 * @returns {boolean***REMOVED*** true if the compilation had a warning
	 */
	hasWarnings() {
		return (
			this.compilation.getWarnings().length > 0 ||
			this.compilation.children.some(child => child.getStats().hasWarnings())
		);
	***REMOVED***

	/**
	 * @returns {boolean***REMOVED*** true if the compilation encountered an error
	 */
	hasErrors() {
		return (
			this.compilation.errors.length > 0 ||
			this.compilation.children.some(child => child.getStats().hasErrors())
		);
	***REMOVED***

	/**
	 * @param {(string | boolean | StatsOptions)=***REMOVED*** options stats options
	 * @returns {StatsCompilation***REMOVED*** json output
	 */
	toJson(options) {
		const normalizedOptions = this.compilation.createStatsOptions(options, {
			forToString: false
		***REMOVED***);

		const statsFactory = this.compilation.createStatsFactory(normalizedOptions);

		return statsFactory.create("compilation", this.compilation, {
			compilation: this.compilation
		***REMOVED***);
	***REMOVED***

	/**
	 * @param {(string | boolean | StatsOptions)=***REMOVED*** options stats options
	 * @returns {string***REMOVED*** string output
	 */
	toString(options) {
		const normalizedOptions = this.compilation.createStatsOptions(options, {
			forToString: true
		***REMOVED***);

		const statsFactory = this.compilation.createStatsFactory(normalizedOptions);
		const statsPrinter = this.compilation.createStatsPrinter(normalizedOptions);

		const data = statsFactory.create("compilation", this.compilation, {
			compilation: this.compilation
		***REMOVED***);
		const result = statsPrinter.print("compilation", data);
		return result === undefined ? "" : result;
	***REMOVED***
***REMOVED***

module.exports = Stats;
