// webpack/lib/Watching.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const Stats = require("./Stats");

/** @typedef {import("../declarations/WebpackOptions").WatchOptions***REMOVED*** WatchOptions */
/** @typedef {import("./Compilation")***REMOVED*** Compilation */
/** @typedef {import("./Compiler")***REMOVED*** Compiler */
/** @typedef {import("./FileSystemInfo").FileSystemInfoEntry***REMOVED*** FileSystemInfoEntry */
/** @typedef {import("./WebpackError")***REMOVED*** WebpackError */
/** @typedef {import("./logging/Logger").Logger***REMOVED*** Logger */
/** @typedef {import("./util/fs").TimeInfoEntries***REMOVED*** TimeInfoEntries */
/** @typedef {import("./util/fs").WatchFileSystem***REMOVED*** WatchFileSystem */
/** @typedef {import("./util/fs").Watcher***REMOVED*** Watcher */

/**
 * @template T
 * @callback Callback
 * @param {Error | null***REMOVED*** err
 * @param {T=***REMOVED*** result
 */

class Watching {
	/**
	 * @param {Compiler***REMOVED*** compiler the compiler
	 * @param {WatchOptions***REMOVED*** watchOptions options
	 * @param {Callback<Stats>***REMOVED*** handler completion handler
	 */
	constructor(compiler, watchOptions, handler) {
		this.startTime = null;
		this.invalid = false;
		this.handler = handler;
		/** @type {Callback<void>[]***REMOVED*** */
		this.callbacks = [];
		/** @type {Callback<void>[] | undefined***REMOVED*** */
		this._closeCallbacks = undefined;
		this.closed = false;
		this.suspended = false;
		this.blocked = false;
		this._isBlocked = () => false;
		this._onChange = () => {***REMOVED***;
		this._onInvalid = () => {***REMOVED***;
		if (typeof watchOptions === "number") {
			/** @type {WatchOptions***REMOVED*** */
			this.watchOptions = {
				aggregateTimeout: watchOptions
			***REMOVED***;
		***REMOVED*** else if (watchOptions && typeof watchOptions === "object") {
			/** @type {WatchOptions***REMOVED*** */
			this.watchOptions = { ...watchOptions ***REMOVED***;
		***REMOVED*** else {
			/** @type {WatchOptions***REMOVED*** */
			this.watchOptions = {***REMOVED***;
		***REMOVED***
		if (typeof this.watchOptions.aggregateTimeout !== "number") {
			this.watchOptions.aggregateTimeout = 20;
		***REMOVED***
		this.compiler = compiler;
		this.running = false;
		this._initial = true;
		this._invalidReported = true;
		this._needRecords = true;
		this.watcher = undefined;
		this.pausedWatcher = undefined;
		/** @type {Set<string> | undefined***REMOVED*** */
		this._collectedChangedFiles = undefined;
		/** @type {Set<string> | undefined***REMOVED*** */
		this._collectedRemovedFiles = undefined;
		this._done = this._done.bind(this);
		process.nextTick(() => {
			if (this._initial) this._invalidate();
		***REMOVED***);
	***REMOVED***

	/**
	 * @param {ReadonlySet<string> | undefined | null***REMOVED*** changedFiles changed files
	 * @param {ReadonlySet<string> | undefined | null***REMOVED*** removedFiles removed files
	 */
	_mergeWithCollected(changedFiles, removedFiles) {
		if (!changedFiles) return;
		if (!this._collectedChangedFiles) {
			this._collectedChangedFiles = new Set(changedFiles);
			this._collectedRemovedFiles = new Set(removedFiles);
		***REMOVED*** else {
			for (const file of changedFiles) {
				this._collectedChangedFiles.add(file);
				/** @type {Set<string>***REMOVED*** */
				(this._collectedRemovedFiles).delete(file);
			***REMOVED***
			for (const file of /** @type {ReadonlySet<string>***REMOVED*** */ (removedFiles)) {
				this._collectedChangedFiles.delete(file);
				/** @type {Set<string>***REMOVED*** */
				(this._collectedRemovedFiles).add(file);
			***REMOVED***
		***REMOVED***
	***REMOVED***

	/**
	 * @param {TimeInfoEntries=***REMOVED*** fileTimeInfoEntries info for files
	 * @param {TimeInfoEntries=***REMOVED*** contextTimeInfoEntries info for directories
	 * @param {ReadonlySet<string>=***REMOVED*** changedFiles changed files
	 * @param {ReadonlySet<string>=***REMOVED*** removedFiles removed files
	 * @returns {void***REMOVED***
	 */
	_go(fileTimeInfoEntries, contextTimeInfoEntries, changedFiles, removedFiles) {
		this._initial = false;
		if (this.startTime === null) this.startTime = Date.now();
		this.running = true;
		if (this.watcher) {
			this.pausedWatcher = this.watcher;
			this.lastWatcherStartTime = Date.now();
			this.watcher.pause();
			this.watcher = null;
		***REMOVED*** else if (!this.lastWatcherStartTime) {
			this.lastWatcherStartTime = Date.now();
		***REMOVED***
		this.compiler.fsStartTime = Date.now();
		if (
			changedFiles &&
			removedFiles &&
			fileTimeInfoEntries &&
			contextTimeInfoEntries
		) {
			this._mergeWithCollected(changedFiles, removedFiles);
			this.compiler.fileTimestamps = fileTimeInfoEntries;
			this.compiler.contextTimestamps = contextTimeInfoEntries;
		***REMOVED*** else if (this.pausedWatcher) {
			if (this.pausedWatcher.getInfo) {
				const {
					changes,
					removals,
					fileTimeInfoEntries,
					contextTimeInfoEntries
				***REMOVED*** = this.pausedWatcher.getInfo();
				this._mergeWithCollected(changes, removals);
				this.compiler.fileTimestamps = fileTimeInfoEntries;
				this.compiler.contextTimestamps = contextTimeInfoEntries;
			***REMOVED*** else {
				this._mergeWithCollected(
					this.pausedWatcher.getAggregatedChanges &&
						this.pausedWatcher.getAggregatedChanges(),
					this.pausedWatcher.getAggregatedRemovals &&
						this.pausedWatcher.getAggregatedRemovals()
				);
				this.compiler.fileTimestamps =
					this.pausedWatcher.getFileTimeInfoEntries();
				this.compiler.contextTimestamps =
					this.pausedWatcher.getContextTimeInfoEntries();
			***REMOVED***
		***REMOVED***
		this.compiler.modifiedFiles = this._collectedChangedFiles;
		this._collectedChangedFiles = undefined;
		this.compiler.removedFiles = this._collectedRemovedFiles;
		this._collectedRemovedFiles = undefined;

		const run = () => {
			if (this.compiler.idle) {
				return this.compiler.cache.endIdle(err => {
					if (err) return this._done(err);
					this.compiler.idle = false;
					run();
				***REMOVED***);
			***REMOVED***
			if (this._needRecords) {
				return this.compiler.readRecords(err => {
					if (err) return this._done(err);

					this._needRecords = false;
					run();
				***REMOVED***);
			***REMOVED***
			this.invalid = false;
			this._invalidReported = false;
			this.compiler.hooks.watchRun.callAsync(this.compiler, err => {
				if (err) return this._done(err);
				/**
				 * @param {Error | null***REMOVED*** err error
				 * @param {Compilation=***REMOVED*** _compilation compilation
				 * @returns {void***REMOVED***
				 */
				const onCompiled = (err, _compilation) => {
					if (err) return this._done(err, _compilation);

					const compilation = /** @type {Compilation***REMOVED*** */ (_compilation);

					if (this.invalid) return this._done(null, compilation);

					if (this.compiler.hooks.shouldEmit.call(compilation) === false) {
						return this._done(null, compilation);
					***REMOVED***

					process.nextTick(() => {
						const logger = compilation.getLogger("webpack.Compiler");
						logger.time("emitAssets");
						this.compiler.emitAssets(compilation, err => {
							logger.timeEnd("emitAssets");
							if (err) return this._done(err, compilation);
							if (this.invalid) return this._done(null, compilation);

							logger.time("emitRecords");
							this.compiler.emitRecords(err => {
								logger.timeEnd("emitRecords");
								if (err) return this._done(err, compilation);

								if (compilation.hooks.needAdditionalPass.call()) {
									compilation.needAdditionalPass = true;

									compilation.startTime = /** @type {number***REMOVED*** */ (
										this.startTime
									);
									compilation.endTime = Date.now();
									logger.time("done hook");
									const stats = new Stats(compilation);
									this.compiler.hooks.done.callAsync(stats, err => {
										logger.timeEnd("done hook");
										if (err) return this._done(err, compilation);

										this.compiler.hooks.additionalPass.callAsync(err => {
											if (err) return this._done(err, compilation);
											this.compiler.compile(onCompiled);
										***REMOVED***);
									***REMOVED***);
									return;
								***REMOVED***
								return this._done(null, compilation);
							***REMOVED***);
						***REMOVED***);
					***REMOVED***);
				***REMOVED***;
				this.compiler.compile(onCompiled);
			***REMOVED***);
		***REMOVED***;

		run();
	***REMOVED***

	/**
	 * @param {Compilation***REMOVED*** compilation the compilation
	 * @returns {Stats***REMOVED*** the compilation stats
	 */
	_getStats(compilation) {
		const stats = new Stats(compilation);
		return stats;
	***REMOVED***

	/**
	 * @param {(Error | null)=***REMOVED*** err an optional error
	 * @param {Compilation=***REMOVED*** compilation the compilation
	 * @returns {void***REMOVED***
	 */
	_done(err, compilation) {
		this.running = false;

		const logger =
			/** @type {Logger***REMOVED*** */
			(compilation && compilation.getLogger("webpack.Watching"));

		/** @type {Stats | undefined***REMOVED*** */
		let stats;

		/**
		 * @param {Error***REMOVED*** err error
		 * @param {Callback<void>[]=***REMOVED*** cbs callbacks
		 */
		const handleError = (err, cbs) => {
			this.compiler.hooks.failed.call(err);
			this.compiler.cache.beginIdle();
			this.compiler.idle = true;
			this.handler(err, /** @type {Stats***REMOVED*** */ (stats));
			if (!cbs) {
				cbs = this.callbacks;
				this.callbacks = [];
			***REMOVED***
			for (const cb of cbs) cb(err);
		***REMOVED***;

		if (
			this.invalid &&
			!this.suspended &&
			!this.blocked &&
			!(this._isBlocked() && (this.blocked = true))
		) {
			if (compilation) {
				logger.time("storeBuildDependencies");
				this.compiler.cache.storeBuildDependencies(
					compilation.buildDependencies,
					err => {
						logger.timeEnd("storeBuildDependencies");
						if (err) return handleError(err);
						this._go();
					***REMOVED***
				);
			***REMOVED*** else {
				this._go();
			***REMOVED***
			return;
		***REMOVED***

		if (compilation) {
			compilation.startTime = /** @type {number***REMOVED*** */ (this.startTime);
			compilation.endTime = Date.now();
			stats = new Stats(compilation);
		***REMOVED***
		this.startTime = null;
		if (err) return handleError(err);

		const cbs = this.callbacks;
		this.callbacks = [];
		logger.time("done hook");
		this.compiler.hooks.done.callAsync(/** @type {Stats***REMOVED*** */ (stats), err => {
			logger.timeEnd("done hook");
			if (err) return handleError(err, cbs);
			this.handler(null, stats);
			logger.time("storeBuildDependencies");
			this.compiler.cache.storeBuildDependencies(
				/** @type {Compilation***REMOVED*** */
				(compilation).buildDependencies,
				err => {
					logger.timeEnd("storeBuildDependencies");
					if (err) return handleError(err, cbs);
					logger.time("beginIdle");
					this.compiler.cache.beginIdle();
					this.compiler.idle = true;
					logger.timeEnd("beginIdle");
					process.nextTick(() => {
						if (!this.closed) {
							this.watch(
								/** @type {Compilation***REMOVED*** */
								(compilation).fileDependencies,
								/** @type {Compilation***REMOVED*** */
								(compilation).contextDependencies,
								/** @type {Compilation***REMOVED*** */
								(compilation).missingDependencies
							);
						***REMOVED***
					***REMOVED***);
					for (const cb of cbs) cb(null);
					this.compiler.hooks.afterDone.call(/** @type {Stats***REMOVED*** */ (stats));
				***REMOVED***
			);
		***REMOVED***);
	***REMOVED***

	/**
	 * @param {Iterable<string>***REMOVED*** files watched files
	 * @param {Iterable<string>***REMOVED*** dirs watched directories
	 * @param {Iterable<string>***REMOVED*** missing watched existence entries
	 * @returns {void***REMOVED***
	 */
	watch(files, dirs, missing) {
		this.pausedWatcher = null;
		this.watcher =
			/** @type {WatchFileSystem***REMOVED*** */
			(this.compiler.watchFileSystem).watch(
				files,
				dirs,
				missing,
				/** @type {number***REMOVED*** */ (this.lastWatcherStartTime),
				this.watchOptions,
				(
					err,
					fileTimeInfoEntries,
					contextTimeInfoEntries,
					changedFiles,
					removedFiles
				) => {
					if (err) {
						this.compiler.modifiedFiles = undefined;
						this.compiler.removedFiles = undefined;
						this.compiler.fileTimestamps = undefined;
						this.compiler.contextTimestamps = undefined;
						this.compiler.fsStartTime = undefined;
						return this.handler(err);
					***REMOVED***
					this._invalidate(
						fileTimeInfoEntries,
						contextTimeInfoEntries,
						changedFiles,
						removedFiles
					);
					this._onChange();
				***REMOVED***,
				(fileName, changeTime) => {
					if (!this._invalidReported) {
						this._invalidReported = true;
						this.compiler.hooks.invalid.call(fileName, changeTime);
					***REMOVED***
					this._onInvalid();
				***REMOVED***
			);
	***REMOVED***

	/**
	 * @param {Callback<void>=***REMOVED*** callback signals when the build has completed again
	 * @returns {void***REMOVED***
	 */
	invalidate(callback) {
		if (callback) {
			this.callbacks.push(callback);
		***REMOVED***
		if (!this._invalidReported) {
			this._invalidReported = true;
			this.compiler.hooks.invalid.call(null, Date.now());
		***REMOVED***
		this._onChange();
		this._invalidate();
	***REMOVED***

	/**
	 * @param {TimeInfoEntries=***REMOVED*** fileTimeInfoEntries info for files
	 * @param {TimeInfoEntries=***REMOVED*** contextTimeInfoEntries info for directories
	 * @param {ReadonlySet<string>=***REMOVED*** changedFiles changed files
	 * @param {ReadonlySet<string>=***REMOVED*** removedFiles removed files
	 * @returns {void***REMOVED***
	 */
	_invalidate(
		fileTimeInfoEntries,
		contextTimeInfoEntries,
		changedFiles,
		removedFiles
	) {
		if (this.suspended || (this._isBlocked() && (this.blocked = true))) {
			this._mergeWithCollected(changedFiles, removedFiles);
			return;
		***REMOVED***

		if (this.running) {
			this._mergeWithCollected(changedFiles, removedFiles);
			this.invalid = true;
		***REMOVED*** else {
			this._go(
				fileTimeInfoEntries,
				contextTimeInfoEntries,
				changedFiles,
				removedFiles
			);
		***REMOVED***
	***REMOVED***

	suspend() {
		this.suspended = true;
	***REMOVED***

	resume() {
		if (this.suspended) {
			this.suspended = false;
			this._invalidate();
		***REMOVED***
	***REMOVED***

	/**
	 * @param {Callback<void>***REMOVED*** callback signals when the watcher is closed
	 * @returns {void***REMOVED***
	 */
	close(callback) {
		if (this._closeCallbacks) {
			if (callback) {
				this._closeCallbacks.push(callback);
			***REMOVED***
			return;
		***REMOVED***
		/**
		 * @param {WebpackError | null***REMOVED*** err error if any
		 * @param {Compilation=***REMOVED*** compilation compilation if any
		 */
		const finalCallback = (err, compilation) => {
			this.running = false;
			this.compiler.running = false;
			this.compiler.watching = undefined;
			this.compiler.watchMode = false;
			this.compiler.modifiedFiles = undefined;
			this.compiler.removedFiles = undefined;
			this.compiler.fileTimestamps = undefined;
			this.compiler.contextTimestamps = undefined;
			this.compiler.fsStartTime = undefined;
			/**
			 * @param {WebpackError | null***REMOVED*** err error if any
			 */
			const shutdown = err => {
				this.compiler.hooks.watchClose.call();
				const closeCallbacks =
					/** @type {Callback<void>[]***REMOVED*** */
					(this._closeCallbacks);
				this._closeCallbacks = undefined;
				for (const cb of closeCallbacks) cb(err);
			***REMOVED***;
			if (compilation) {
				const logger = compilation.getLogger("webpack.Watching");
				logger.time("storeBuildDependencies");
				this.compiler.cache.storeBuildDependencies(
					compilation.buildDependencies,
					err2 => {
						logger.timeEnd("storeBuildDependencies");
						shutdown(err || err2);
					***REMOVED***
				);
			***REMOVED*** else {
				shutdown(err);
			***REMOVED***
		***REMOVED***;

		this.closed = true;
		if (this.watcher) {
			this.watcher.close();
			this.watcher = null;
		***REMOVED***
		if (this.pausedWatcher) {
			this.pausedWatcher.close();
			this.pausedWatcher = null;
		***REMOVED***
		this._closeCallbacks = [];
		if (callback) {
			this._closeCallbacks.push(callback);
		***REMOVED***
		if (this.running) {
			this.invalid = true;
			this._done = finalCallback;
		***REMOVED*** else {
			finalCallback(null);
		***REMOVED***
	***REMOVED***
***REMOVED***

module.exports = Watching;
