// webpack/lib/WebpackError.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Jarid Margolin @jaridmargolin
*/

"use strict";

const inspect = require("util").inspect.custom;
const makeSerializable = require("./util/makeSerializable");

/** @typedef {import("./Chunk")***REMOVED*** Chunk */
/** @typedef {import("./Dependency").DependencyLocation***REMOVED*** DependencyLocation */
/** @typedef {import("./Module")***REMOVED*** Module */
/** @typedef {import("./serialization/ObjectMiddleware").ObjectDeserializerContext***REMOVED*** ObjectDeserializerContext */
/** @typedef {import("./serialization/ObjectMiddleware").ObjectSerializerContext***REMOVED*** ObjectSerializerContext */

class WebpackError extends Error {
	/**
	 * Creates an instance of WebpackError.
	 * @param {string=***REMOVED*** message error message
	 * @param {{ cause?: unknown ***REMOVED******REMOVED*** options error options
	 */
	constructor(message, options = {***REMOVED***) {
		// @ts-expect-error ES2018 doesn't `Error.cause`, but it can be used by developers
		super(message, options);

		/** @type {string=***REMOVED*** */
		this.details = undefined;
		/** @type {(Module | null)=***REMOVED*** */
		this.module = undefined;
		/** @type {DependencyLocation=***REMOVED*** */
		this.loc = undefined;
		/** @type {boolean=***REMOVED*** */
		this.hideStack = undefined;
		/** @type {Chunk=***REMOVED*** */
		this.chunk = undefined;
		/** @type {string=***REMOVED*** */
		this.file = undefined;
	***REMOVED***

	[inspect]() {
		return (
			this.stack +
			(this.details ? `\n${this.details***REMOVED***` : "") +
			(this.cause ? `\n${this.cause***REMOVED***` : "")
		);
	***REMOVED***

	/**
	 * @param {ObjectSerializerContext***REMOVED*** context context
	 */
	serialize({ write ***REMOVED***) {
		write(this.name);
		write(this.message);
		write(this.stack);
		write(this.cause);
		write(this.details);
		write(this.loc);
		write(this.hideStack);
	***REMOVED***

	/**
	 * @param {ObjectDeserializerContext***REMOVED*** context context
	 */
	deserialize({ read ***REMOVED***) {
		this.name = read();
		this.message = read();
		this.stack = read();
		this.cause = read();
		this.details = read();
		this.loc = read();
		this.hideStack = read();
	***REMOVED***
***REMOVED***

makeSerializable(WebpackError, "webpack/lib/WebpackError");

module.exports = WebpackError;
