// react-dev-utils/browsersHelper.js
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
'use strict';

const browserslist = require('browserslist');
const chalk = require('chalk');
const os = require('os');
const prompts = require('prompts');
const pkgUp = require('pkg-up');
const fs = require('fs');

const defaultBrowsers = {
  production: ['>0.2%', 'not dead', 'not op_mini all'],
  development: [
    'last 1 chrome version',
    'last 1 firefox version',
    'last 1 safari version',
  ],
***REMOVED***;

function shouldSetBrowsers(isInteractive) {
  if (!isInteractive) {
    return Promise.resolve(true);
  ***REMOVED***

  const question = {
    type: 'confirm',
    name: 'shouldSetBrowsers',
    message:
      chalk.yellow("We're unable to detect target browsers.") +
      `\n\nWould you like to add the defaults to your ${chalk.bold(
        'package.json'
      )***REMOVED***?`,
    initial: true,
  ***REMOVED***;

  return prompts(question).then(answer => answer.shouldSetBrowsers);
***REMOVED***

function checkBrowsers(dir, isInteractive, retry = true) {
  const current = browserslist.loadConfig({ path: dir ***REMOVED***);
  if (current != null) {
    return Promise.resolve(current);
  ***REMOVED***

  if (!retry) {
    return Promise.reject(
      new Error(
        chalk.red(
          'As of react-scripts >=2 you must specify targeted browsers.'
        ) +
          os.EOL +
          `Please add a ${chalk.underline(
            'browserslist'
          )***REMOVED*** key to your ${chalk.bold('package.json')***REMOVED***.`
      )
    );
  ***REMOVED***

  return shouldSetBrowsers(isInteractive).then(shouldSetBrowsers => {
    if (!shouldSetBrowsers) {
      return checkBrowsers(dir, isInteractive, false);
    ***REMOVED***

    return (
      pkgUp({ cwd: dir ***REMOVED***)
        .then(filePath => {
          if (filePath == null) {
            return Promise.reject();
          ***REMOVED***
          const pkg = JSON.parse(fs.readFileSync(filePath));
          pkg['browserslist'] = defaultBrowsers;
          fs.writeFileSync(filePath, JSON.stringify(pkg, null, 2) + os.EOL);

          browserslist.clearCaches();
          console.log();
          console.log(
            `${chalk.green('Set target browsers:')***REMOVED*** ${chalk.cyan(
              defaultBrowsers.join(', ')
            )***REMOVED***`
          );
          console.log();
        ***REMOVED***)
        // Swallow any error
        .catch(() => {***REMOVED***)
        .then(() => checkBrowsers(dir, isInteractive, false))
    );
  ***REMOVED***);
***REMOVED***

module.exports = { defaultBrowsers, checkBrowsers ***REMOVED***;
