// react-dev-utils/ModuleNotFoundPlugin.js
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

const chalk = require('chalk');
const findUp = require('find-up');
const path = require('path');

class ModuleNotFoundPlugin {
  constructor(appPath, yarnLockFile) {
    this.appPath = appPath;
    this.yarnLockFile = yarnLockFile;

    this.useYarnCommand = this.useYarnCommand.bind(this);
    this.getRelativePath = this.getRelativePath.bind(this);
    this.prettierError = this.prettierError.bind(this);
  ***REMOVED***

  useYarnCommand() {
    try {
      return findUp.sync('yarn.lock', { cwd: this.appPath ***REMOVED***) != null;
    ***REMOVED*** catch (_) {
      return false;
    ***REMOVED***
  ***REMOVED***

  getRelativePath(_file) {
    let file = path.relative(this.appPath, _file);
    if (file.startsWith('..')) {
      file = _file;
    ***REMOVED*** else if (!file.startsWith('.')) {
      file = '.' + path.sep + file;
    ***REMOVED***
    return file;
  ***REMOVED***

  prettierError(err) {
    let { details: _details = '', origin ***REMOVED*** = err;

    if (origin == null) {
      const caseSensitivity =
        err.message &&
        /\[CaseSensitivePathsPlugin\] `(.*?)` .* `(.*?)`/.exec(err.message);
      if (caseSensitivity) {
        const [, incorrectPath, actualName] = caseSensitivity;
        const actualFile = this.getRelativePath(
          path.join(path.dirname(incorrectPath), actualName)
        );
        const incorrectName = path.basename(incorrectPath);
        err.message = `Cannot find file: '${incorrectName***REMOVED***' does not match the corresponding name on disk: '${actualFile***REMOVED***'.`;
      ***REMOVED***
      return err;
    ***REMOVED***

    const file = this.getRelativePath(origin.resource);
    let details = _details.split('\n');

    const request = /resolve '(.*?)' in '(.*?)'/.exec(details);
    if (request) {
      const isModule = details[1] && details[1].includes('module');
      const isFile = details[1] && details[1].includes('file');

      let [, target, context] = request;
      context = this.getRelativePath(context);
      if (isModule) {
        const isYarn = this.useYarnCommand();
        details = [
          `Cannot find module: '${target***REMOVED***'. Make sure this package is installed.`,
          '',
          'You can install this package by running: ' +
            (isYarn
              ? chalk.bold(`yarn add ${target***REMOVED***`)
              : chalk.bold(`npm install ${target***REMOVED***`)) +
            '.',
        ];
      ***REMOVED*** else if (isFile) {
        details = [`Cannot find file '${target***REMOVED***' in '${context***REMOVED***'.`];
      ***REMOVED*** else {
        details = [err.message];
      ***REMOVED***
    ***REMOVED*** else {
      details = [err.message];
    ***REMOVED***
    err.message = [file, ...details].join('\n').replace('Error: ', '');

    const isModuleScopePluginError =
      err.error && err.error.__module_scope_plugin;
    if (isModuleScopePluginError) {
      err.message = err.message.replace('Module not found: ', '');
    ***REMOVED***
    return err;
  ***REMOVED***

  apply(compiler) {
    const { prettierError ***REMOVED*** = this;
    compiler.hooks.make.intercept({
      register(tap) {
        if (!(tap.name === 'MultiEntryPlugin' || tap.name === 'EntryPlugin')) {
          return tap;
        ***REMOVED***
        return Object.assign({***REMOVED***, tap, {
          fn: (compilation, callback) => {
            tap.fn(compilation, (err, ...args) => {
              if (err && err.name === 'ModuleNotFoundError') {
                err = prettierError(err);
              ***REMOVED***
              callback(err, ...args);
            ***REMOVED***);
          ***REMOVED***,
        ***REMOVED***);
      ***REMOVED***,
    ***REMOVED***);
    compiler.hooks.normalModuleFactory.tap('ModuleNotFoundPlugin', nmf => {
      nmf.hooks.afterResolve.intercept({
        register(tap) {
          if (tap.name !== 'CaseSensitivePathsPlugin') {
            return tap;
          ***REMOVED***
          return Object.assign({***REMOVED***, tap, {
            fn: (compilation, callback) => {
              tap.fn(compilation, (err, ...args) => {
                if (
                  err &&
                  err.message &&
                  err.message.includes('CaseSensitivePathsPlugin')
                ) {
                  err = prettierError(err);
                ***REMOVED***
                callback(err, ...args);
              ***REMOVED***);
            ***REMOVED***,
          ***REMOVED***);
        ***REMOVED***,
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***
***REMOVED***

module.exports = ModuleNotFoundPlugin;
