// locate-path/index.js
'use strict';
const path = require('path');
const fs = require('fs');
const {promisify***REMOVED*** = require('util');
const pLocate = require('p-locate');

const fsStat = promisify(fs.stat);
const fsLStat = promisify(fs.lstat);

const typeMappings = {
	directory: 'isDirectory',
	file: 'isFile'
***REMOVED***;

function checkType({type***REMOVED***) {
	if (type in typeMappings) {
		return;
	***REMOVED***

	throw new Error(`Invalid type specified: ${type***REMOVED***`);
***REMOVED***

const matchType = (type, stat) => type === undefined || stat[typeMappings[type]]();

module.exports = async (paths, options) => {
	options = {
		cwd: process.cwd(),
		type: 'file',
		allowSymlinks: true,
		...options
	***REMOVED***;

	checkType(options);

	const statFn = options.allowSymlinks ? fsStat : fsLStat;

	return pLocate(paths, async path_ => {
		try {
			const stat = await statFn(path.resolve(options.cwd, path_));
			return matchType(options.type, stat);
		***REMOVED*** catch {
			return false;
		***REMOVED***
	***REMOVED***, options);
***REMOVED***;

module.exports.sync = (paths, options) => {
	options = {
		cwd: process.cwd(),
		allowSymlinks: true,
		type: 'file',
		...options
	***REMOVED***;

	checkType(options);

	const statFn = options.allowSymlinks ? fs.statSync : fs.lstatSync;

	for (const path_ of paths) {
		try {
			const stat = statFn(path.resolve(options.cwd, path_));

			if (matchType(options.type, stat)) {
				return path_;
			***REMOVED***
		***REMOVED*** catch {***REMOVED***
	***REMOVED***
***REMOVED***;
