// mz/readline.js
var readline = require('readline')
var Promise = require('any-promise')
var objectAssign = require('object-assign')
var Interface = readline.Interface

function wrapCompleter (completer) {
  if (completer.length === 2) return completer

  return function (line, cb) {
    var result = completer(line)

    if (typeof result.then !== 'function') {
      return cb(null, result)
    ***REMOVED***

    result.catch(cb).then(function (result) {
      process.nextTick(function () { cb(null, result) ***REMOVED***)
    ***REMOVED***)
  ***REMOVED***
***REMOVED***

function InterfaceAsPromised (input, output, completer, terminal) {
  if (arguments.length === 1) {
    var options = input

    if (typeof options.completer === 'function') {
      options = objectAssign({***REMOVED***, options, {
        completer: wrapCompleter(options.completer)
      ***REMOVED***)
    ***REMOVED***

    Interface.call(this, options)
  ***REMOVED*** else {
    if (typeof completer === 'function') {
      completer = wrapCompleter(completer)
    ***REMOVED***

    Interface.call(this, input, output, completer, terminal)
  ***REMOVED***
***REMOVED***

InterfaceAsPromised.prototype = Object.create(Interface.prototype)

InterfaceAsPromised.prototype.question = function (question, callback) {
  if (typeof callback === 'function') {
    return Interface.prototype.question.call(this, question, callback)
  ***REMOVED***

  var self = this
  return new Promise(function (resolve) {
    Interface.prototype.question.call(self, question, resolve)
  ***REMOVED***)
***REMOVED***

objectAssign(exports, readline, {
  Interface: InterfaceAsPromised,
  createInterface: function (input, output, completer, terminal) {
    if (arguments.length === 1) {
      return new InterfaceAsPromised(input)
    ***REMOVED***

    return new InterfaceAsPromised(input, output, completer, terminal)
  ***REMOVED***
***REMOVED***)
