# Enforce assertion to be made in a test body (`expect-expect`)

Ensure that there is at least one `expect` call made in a test.

## Rule details

This rule triggers when there is no call made to `expect` in a test, to prevent
users from forgetting to add assertions.

Examples of **incorrect** code for this rule:

```js
it('should be a test', () => {
  console.log('no assertion');
***REMOVED***);
test('should assert something', () => {***REMOVED***);
```

Examples of **correct** code for this rule:

```js
it('should be a test', () => {
  expect(true).toBeDefined();
***REMOVED***);
it('should work with callbacks/async', () => {
  somePromise().then(res => expect(res).toBe('passed'));
***REMOVED***);
```

## Options

```json
{
  "jest/expect-expect": [
    "error",
    {
      "assertFunctionNames": ["expect"],
      "additionalTestBlockFunctions": []
    ***REMOVED***
  ]
***REMOVED***
```

### `assertFunctionNames`

This array option specifies the names of functions that should be considered to
be asserting functions. Function names can use wildcards i.e `request.*.expect`,
`request.**.expect`, `request.*.expect*`

Examples of **incorrect** code for the `{ "assertFunctionNames": ["expect"] ***REMOVED***`
option:

```js
/* eslint jest/expect-expect: ["error", { "assertFunctionNames": ["expect"] ***REMOVED***] */

import { expectSaga ***REMOVED*** from 'redux-saga-test-plan';
import { addSaga ***REMOVED*** from '../src/sagas';

test('returns sum', () => {
  expectSaga(addSaga, 1, 1).returns(2).run();
***REMOVED***);
```

Examples of **correct** code for the
`{ "assertFunctionNames": ["expect", "expectSaga"] ***REMOVED***` option:

```js
/* eslint jest/expect-expect: ["error", { "assertFunctionNames": ["expect", "expectSaga"] ***REMOVED***] */

import { expectSaga ***REMOVED*** from 'redux-saga-test-plan';
import { addSaga ***REMOVED*** from '../src/sagas';

test('returns sum', () => {
  expectSaga(addSaga, 1, 1).returns(2).run();
***REMOVED***);
```

Since the string is compiled into a regular expression, you'll need to escape
special characters such as `$` with a double backslash:

```js
/* eslint jest/expect-expect: ["error", { "assertFunctionNames": ["expect\\$"] ***REMOVED***] */

it('is money-like', () => {
  expect$(1.0);
***REMOVED***);
```

Examples of **correct** code for working with the HTTP assertions library
[SuperTest](https://www.npmjs.com/package/supertest) with the
`{ "assertFunctionNames": ["expect", "request.**.expect"] ***REMOVED***` option:

```js
/* eslint jest/expect-expect: ["error", { "assertFunctionNames": ["expect", "request.**.expect"] ***REMOVED***] */
const request = require('supertest');
const express = require('express');

const app = express();

describe('GET /user', function () {
  it('responds with json', function (done) {
    request(app).get('/user').expect('Content-Type', /json/).expect(200, done);
  ***REMOVED***);
***REMOVED***);
```

### `additionalTestBlockFunctions`

This array can be used to specify the names of functions that should also be
treated as test blocks:

```json
{
  "rules": {
    "jest/expect-expect": [
      "error",
      { "additionalTestBlockFunctions": ["theoretically"] ***REMOVED***
    ]
  ***REMOVED***
***REMOVED***
```

The following is _correct_ when using the above configuration:

```js
import theoretically from 'jest-theories';

describe('NumberToLongString', () => {
  const theories = [
    { input: 100, expected: 'One hundred' ***REMOVED***,
    { input: 1000, expected: 'One thousand' ***REMOVED***,
    { input: 10000, expected: 'Ten thousand' ***REMOVED***,
    { input: 100000, expected: 'One hundred thousand' ***REMOVED***,
  ];

  theoretically(
    'the number {input***REMOVED*** is correctly translated to string',
    theories,
    theory => {
      const output = NumberToLongString(theory.input);
      expect(output).toBe(theory.expected);
    ***REMOVED***,
  );
***REMOVED***);
```
