# Disallow duplicate setup and teardown hooks (`no-duplicate-hooks`)

A `describe` block should not contain duplicate hooks.

## Rule Details

Examples of **incorrect** code for this rule

```js
/* eslint jest/no-duplicate-hooks: "error" */

describe('foo', () => {
  beforeEach(() => {
    // some setup
  ***REMOVED***);
  beforeEach(() => {
    // some setup
  ***REMOVED***);
  test('foo_test', () => {
    // some test
  ***REMOVED***);
***REMOVED***);

// Nested describe scenario
describe('foo', () => {
  beforeEach(() => {
    // some setup
  ***REMOVED***);
  test('foo_test', () => {
    // some test
  ***REMOVED***);
  describe('bar', () => {
    test('bar_test', () => {
      afterAll(() => {
        // some teardown
      ***REMOVED***);
      afterAll(() => {
        // some teardown
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***);
```

Examples of **correct** code for this rule

```js
/* eslint jest/no-duplicate-hooks: "error" */

describe('foo', () => {
  beforeEach(() => {
    // some setup
  ***REMOVED***);
  test('foo_test', () => {
    // some test
  ***REMOVED***);
***REMOVED***);

// Nested describe scenario
describe('foo', () => {
  beforeEach(() => {
    // some setup
  ***REMOVED***);
  test('foo_test', () => {
    // some test
  ***REMOVED***);
  describe('bar', () => {
    test('bar_test', () => {
      beforeEach(() => {
        // some setup
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***);
```
