# Disallow setup and teardown hooks (`no-hooks`)

Jest provides global functions for setup and teardown tasks, which are called
before/after each test case and each test suite. The use of these hooks promotes
shared state between tests.

## Rule Details

This rule reports for the following function calls:

- `beforeAll`
- `beforeEach`
- `afterAll`
- `afterEach`

Examples of **incorrect** code for this rule:

```js
/* eslint jest/no-hooks: "error" */

function setupFoo(options) {
  /* ... */
***REMOVED***

function setupBar(options) {
  /* ... */
***REMOVED***

describe('foo', () => {
  let foo;

  beforeEach(() => {
    foo = setupFoo();
  ***REMOVED***);

  afterEach(() => {
    foo = null;
  ***REMOVED***);

  it('does something', () => {
    expect(foo.doesSomething()).toBe(true);
  ***REMOVED***);

  describe('with bar', () => {
    let bar;

    beforeEach(() => {
      bar = setupBar();
    ***REMOVED***);

    afterEach(() => {
      bar = null;
    ***REMOVED***);

    it('does something with bar', () => {
      expect(foo.doesSomething(bar)).toBe(true);
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***);
```

Examples of **correct** code for this rule:

```js
/* eslint jest/no-hooks: "error" */

function setupFoo(options) {
  /* ... */
***REMOVED***

function setupBar(options) {
  /* ... */
***REMOVED***

describe('foo', () => {
  it('does something', () => {
    const foo = setupFoo();
    expect(foo.doesSomething()).toBe(true);
  ***REMOVED***);

  it('does something with bar', () => {
    const foo = setupFoo();
    const bar = setupBar();
    expect(foo.doesSomething(bar)).toBe(true);
  ***REMOVED***);
***REMOVED***);
```

## Options

```json
{
  "jest/no-hooks": [
    "error",
    {
      "allow": ["afterEach", "afterAll"]
    ***REMOVED***
  ]
***REMOVED***
```

### `allow`

This array option controls which Jest hooks are checked by this rule. There are
four possible values:

- `"beforeAll"`
- `"beforeEach"`
- `"afterAll"`
- `"afterEach"`

By default, none of these options are enabled (the equivalent of
`{ "allow": [] ***REMOVED***`).

Examples of **incorrect** code for the `{ "allow": ["afterEach"] ***REMOVED***` option:

```js
/* eslint jest/no-hooks: ["error", { "allow": ["afterEach"] ***REMOVED***] */

function setupFoo(options) {
  /* ... */
***REMOVED***

let foo;

beforeEach(() => {
  foo = setupFoo();
***REMOVED***);

afterEach(() => {
  jest.resetModules();
***REMOVED***);

test('foo does this', () => {
  // ...
***REMOVED***);

test('foo does that', () => {
  // ...
***REMOVED***);
```

Examples of **correct** code for the `{ "allow": ["afterEach"] ***REMOVED***` option:

```js
/* eslint jest/no-hooks: ["error", { "allow": ["afterEach"] ***REMOVED***] */

function setupFoo(options) {
  /* ... */
***REMOVED***

afterEach(() => {
  jest.resetModules();
***REMOVED***);

test('foo does this', () => {
  const foo = setupFoo();
  // ...
***REMOVED***);

test('foo does that', () => {
  const foo = setupFoo();
  // ...
***REMOVED***);
```

## When Not To Use It

If you prefer using the setup and teardown hooks provided by Jest, you can
safely disable this rule.

## Further Reading

- [Jest docs - Setup and Teardown](https://facebook.github.io/jest/docs/en/setup-teardown.html)
