# Suggest having hooks before any test cases (`prefer-hooks-on-top`)

While hooks can be setup anywhere in a test file, they are always called in a
specific order which means it can be confusing if they're intermixed with test
cases.

This rule helps to ensure that hooks are always defined before test cases.

## Rule Details

Examples of **incorrect** code for this rule

```js
/* eslint jest/prefer-hooks-on-top: "error" */

describe('foo', () => {
  beforeEach(() => {
    seedMyDatabase();
  ***REMOVED***);

  it('accepts this input', () => {
    // ...
  ***REMOVED***);

  beforeAll(() => {
    createMyDatabase();
  ***REMOVED***);

  it('returns that value', () => {
    // ...
  ***REMOVED***);

  describe('when the database has specific values', () => {
    const specificValue = '...';

    beforeEach(() => {
      seedMyDatabase(specificValue);
    ***REMOVED***);

    it('accepts that input', () => {
      // ...
    ***REMOVED***);

    it('throws an error', () => {
      // ...
    ***REMOVED***);

    afterEach(() => {
      clearLogger();
    ***REMOVED***);
    beforeEach(() => {
      mockLogger();
    ***REMOVED***);

    it('logs a message', () => {
      // ...
    ***REMOVED***);
  ***REMOVED***);

  afterAll(() => {
    removeMyDatabase();
  ***REMOVED***);
***REMOVED***);
```

Examples of **correct** code for this rule

```js
/* eslint jest/prefer-hooks-on-top: "error" */

describe('foo', () => {
  beforeAll(() => {
    createMyDatabase();
  ***REMOVED***);

  beforeEach(() => {
    seedMyDatabase();
  ***REMOVED***);

  afterAll(() => {
    clearMyDatabase();
  ***REMOVED***);

  it('accepts this input', () => {
    // ...
  ***REMOVED***);

  it('returns that value', () => {
    // ...
  ***REMOVED***);

  describe('when the database has specific values', () => {
    const specificValue = '...';

    beforeEach(() => {
      seedMyDatabase(specificValue);
    ***REMOVED***);

    beforeEach(() => {
      mockLogger();
    ***REMOVED***);

    afterEach(() => {
      clearLogger();
    ***REMOVED***);

    it('accepts that input', () => {
      // ...
    ***REMOVED***);

    it('throws an error', () => {
      // ...
    ***REMOVED***);

    it('logs a message', () => {
      // ...
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***);
```
