# Require test cases and hooks to be inside a `describe` block (`require-top-level-describe`)

Jest allows you to organise your test files the way you want it. However, the
more your codebase grows, the more it becomes hard to navigate in your test
files. This rule makes sure you provide at least a top-level `describe` block in
your test file.

## Rule Details

This rule triggers a warning if a test case (`test` and `it`) or a hook
(`beforeAll`, `beforeEach`, `afterEach`, `afterAll`) is not located in a
top-level `describe` block.

The following patterns are considered warnings:

```js
// Above a describe block
test('my test', () => {***REMOVED***);
describe('test suite', () => {
  it('test', () => {***REMOVED***);
***REMOVED***);

// Below a describe block
describe('test suite', () => {***REMOVED***);
test('my test', () => {***REMOVED***);

// Same for hooks
beforeAll('my beforeAll', () => {***REMOVED***);
describe('test suite', () => {***REMOVED***);
afterEach('my afterEach', () => {***REMOVED***);
```

The following patterns are **not** considered warnings:

```js
// In a describe block
describe('test suite', () => {
  test('my test', () => {***REMOVED***);
***REMOVED***);

// In a nested describe block
describe('test suite', () => {
  test('my test', () => {***REMOVED***);
  describe('another test suite', () => {
    test('my other test', () => {***REMOVED***);
  ***REMOVED***);
***REMOVED***);
```

You can also enforce a limit on the number of describes allowed at the top-level
using the `maxNumberOfTopLevelDescribes` option:

```json
{
  "jest/require-top-level-describe": [
    "error",
    {
      "maxNumberOfTopLevelDescribes": 2
    ***REMOVED***
  ]
***REMOVED***
```

Examples of **incorrect** code with the above config:

```js
describe('test suite', () => {
  it('test', () => {***REMOVED***);
***REMOVED***);

describe('test suite', () => {***REMOVED***);

describe('test suite', () => {***REMOVED***);
```

This option defaults to `Infinity`, allowing any number of top-level describes.

## When Not To Use It

Don't use this rule on non-jest test files.
