// postcss-ordered-values/src/rules/animation.js
'use strict';
const { unit ***REMOVED*** = require('postcss-value-parser');
const { getArguments ***REMOVED*** = require('cssnano-utils');
const addSpace = require('../lib/addSpace');
const getValue = require('../lib/getValue');

// animation: [ none | <keyframes-name> ] || <time> || <single-timing-function> || <time> || <single-animation-iteration-count> || <single-animation-direction> || <single-animation-fill-mode> || <single-animation-play-state>
const functions = new Set(['steps', 'cubic-bezier', 'frames']);
const keywords = new Set([
  'ease',
  'ease-in',
  'ease-in-out',
  'ease-out',
  'linear',
  'step-end',
  'step-start',
]);

const directions = new Set([
  'normal',
  'reverse',
  'alternate',
  'alternate-reverse',
]);
const fillModes = new Set(['none', 'forwards', 'backwards', 'both']);
const playStates = new Set(['running', 'paused']);
const timeUnits = new Set(['ms', 's']);

/**
 * @param {string***REMOVED*** value
 * @param {string***REMOVED*** type
 * @return {boolean***REMOVED***
 */
const isTimingFunction = (value, type) => {
  return (type === 'function' && functions.has(value)) || keywords.has(value);
***REMOVED***;
/**
 * @param {string***REMOVED*** value
 * @return {boolean***REMOVED***
 */
const isDirection = (value) => {
  return directions.has(value);
***REMOVED***;
/**
 * @param {string***REMOVED*** value
 * @return {boolean***REMOVED***
 */
const isFillMode = (value) => {
  return fillModes.has(value);
***REMOVED***;
/**
 * @param {string***REMOVED*** value
 * @return {boolean***REMOVED***
 */
const isPlayState = (value) => {
  return playStates.has(value);
***REMOVED***;
/**
 * @param {string***REMOVED*** value
 * @return {boolean***REMOVED***
 */
const isTime = (value) => {
  const quantity = unit(value);

  return quantity && timeUnits.has(quantity.unit);
***REMOVED***;
/**
 * @param {string***REMOVED*** value
 * @return {boolean***REMOVED***
 */
const isIterationCount = (value) => {
  const quantity = unit(value);

  return value === 'infinite' || (quantity && !quantity.unit);
***REMOVED***;

const stateConditions = [
  { property: 'duration', delegate: isTime ***REMOVED***,
  { property: 'timingFunction', delegate: isTimingFunction ***REMOVED***,
  { property: 'delay', delegate: isTime ***REMOVED***,
  { property: 'iterationCount', delegate: isIterationCount ***REMOVED***,
  { property: 'direction', delegate: isDirection ***REMOVED***,
  { property: 'fillMode', delegate: isFillMode ***REMOVED***,
  { property: 'playState', delegate: isPlayState ***REMOVED***,
];
/**
 * @param {import('postcss-value-parser').Node[][]***REMOVED*** args
 * @return {import('postcss-value-parser').Node[][]***REMOVED***
 */
function normalize(args) {
  const list = [];

  for (const arg of args) {
    /** @type {Record<string, import('postcss-value-parser').Node[]>***REMOVED*** */
    const state = {
      name: [],
      duration: [],
      timingFunction: [],
      delay: [],
      iterationCount: [],
      direction: [],
      fillMode: [],
      playState: [],
    ***REMOVED***;

    arg.forEach((node) => {
      let { type, value ***REMOVED*** = node;

      if (type === 'space') {
        return;
      ***REMOVED***

      value = value.toLowerCase();

      const hasMatch = stateConditions.some(({ property, delegate ***REMOVED***) => {
        if (delegate(value, type) && !state[property].length) {
          state[property] = [node, addSpace()];
          return true;
        ***REMOVED***
      ***REMOVED***);

      if (!hasMatch) {
        state.name = [...state.name, node, addSpace()];
      ***REMOVED***
    ***REMOVED***);

    list.push([
      ...state.name,
      ...state.duration,
      ...state.timingFunction,
      ...state.delay,
      ...state.iterationCount,
      ...state.direction,
      ...state.fillMode,
      ...state.playState,
    ]);
  ***REMOVED***
  return list;
***REMOVED***
/**
 * @param {import('postcss-value-parser').ParsedValue***REMOVED*** parsed
 * @return {string***REMOVED***
 */
module.exports = function normalizeAnimation(parsed) {
  const values = normalize(getArguments(parsed));

  return getValue(values);
***REMOVED***;
