// postcss-ordered-values/src/rules/transition.js
'use strict';
const { unit ***REMOVED*** = require('postcss-value-parser');
const { getArguments ***REMOVED*** = require('cssnano-utils');
const addSpace = require('../lib/addSpace');
const getValue = require('../lib/getValue');

// transition: [ none | <single-transition-property> ] || <time> || <single-transition-timing-function> || <time>

const timingFunctions = new Set([
  'ease',
  'linear',
  'ease-in',
  'ease-out',
  'ease-in-out',
  'step-start',
  'step-end',
]);

/**
 * @param {import('postcss-value-parser').Node[][]***REMOVED*** args
 * @return {import('postcss-value-parser').Node[][]***REMOVED***
 */
function normalize(args) {
  const list = [];
  for (const arg of args) {
    /** @type {Record<string, import('postcss-value-parser').Node[]>***REMOVED*** */
    let state = {
      timingFunction: [],
      property: [],
      time1: [],
      time2: [],
    ***REMOVED***;

    arg.forEach((node) => {
      const { type, value ***REMOVED*** = node;

      if (type === 'space') {
        return;
      ***REMOVED***

      if (
        type === 'function' &&
        new Set(['steps', 'cubic-bezier']).has(value.toLowerCase())
      ) {
        state.timingFunction = [...state.timingFunction, node, addSpace()];
      ***REMOVED*** else if (unit(value)) {
        if (!state.time1.length) {
          state.time1 = [...state.time1, node, addSpace()];
        ***REMOVED*** else {
          state.time2 = [...state.time2, node, addSpace()];
        ***REMOVED***
      ***REMOVED*** else if (timingFunctions.has(value.toLowerCase())) {
        state.timingFunction = [...state.timingFunction, node, addSpace()];
      ***REMOVED*** else {
        state.property = [...state.property, node, addSpace()];
      ***REMOVED***
    ***REMOVED***);

    list.push([
      ...state.property,
      ...state.time1,
      ...state.timingFunction,
      ...state.time2,
    ]);
  ***REMOVED***
  return list;
***REMOVED***
/**
 * @param {import('postcss-value-parser').ParsedValue***REMOVED*** parsed
 * @return {string***REMOVED***
 */
module.exports = function normalizeTransition(parsed) {
  const values = normalize(getArguments(parsed));

  return getValue(values);
***REMOVED***;
