// ajv-keywords/src/definitions/deepRequired.ts
import type {CodeKeywordDefinition, KeywordCxt***REMOVED*** from "ajv"
import {_, or, and, getProperty, Code***REMOVED*** from "ajv/dist/compile/codegen"

export default function getDef(): CodeKeywordDefinition {
  return {
    keyword: "deepRequired",
    type: "object",
    schemaType: "array",
    code(ctx: KeywordCxt) {
      const {schema, data***REMOVED*** = ctx
      const props = (schema as string[]).map((jp: string) => _`(${getData(jp)***REMOVED***) === undefined`)
      ctx.fail(or(...props))

      function getData(jsonPointer: string): Code {
        if (jsonPointer === "") throw new Error("empty JSON pointer not allowed")
        const segments = jsonPointer.split("/")
        let x: Code = data
        const xs = segments.map((s, i) =>
          i ? (x = _`${x***REMOVED***${getProperty(unescapeJPSegment(s))***REMOVED***`) : x
        )
        return and(...xs)
      ***REMOVED***
    ***REMOVED***,
    metaSchema: {
      type: "array",
      items: {type: "string", format: "json-pointer"***REMOVED***,
    ***REMOVED***,
  ***REMOVED***
***REMOVED***

function unescapeJPSegment(s: string): string {
  return s.replace(/~1/g, "/").replace(/~0/g, "~")
***REMOVED***

module.exports = getDef
