// ajv-keywords/src/definitions/deepProperties.ts
import type {MacroKeywordDefinition, SchemaObject, Schema***REMOVED*** from "ajv"
import type {DefinitionOptions***REMOVED*** from "./_types"
import {metaSchemaRef***REMOVED*** from "./_util"

export default function getDef(opts?: DefinitionOptions): MacroKeywordDefinition {
  return {
    keyword: "deepProperties",
    type: "object",
    schemaType: "object",
    macro: function (schema: Record<string, SchemaObject>) {
      const allOf = []
      for (const pointer in schema) allOf.push(getSchema(pointer, schema[pointer]))
      return {allOf***REMOVED***
    ***REMOVED***,
    metaSchema: {
      type: "object",
      propertyNames: {type: "string", format: "json-pointer"***REMOVED***,
      additionalProperties: metaSchemaRef(opts),
    ***REMOVED***,
  ***REMOVED***
***REMOVED***

function getSchema(jsonPointer: string, schema: SchemaObject): SchemaObject {
  const segments = jsonPointer.split("/")
  const rootSchema: SchemaObject = {***REMOVED***
  let pointerSchema: SchemaObject = rootSchema
  for (let i = 1; i < segments.length; i++) {
    let segment: string = segments[i]
    const isLast = i === segments.length - 1
    segment = unescapeJsonPointer(segment)
    const properties: Record<string, Schema> = (pointerSchema.properties = {***REMOVED***)
    let items: SchemaObject[] | undefined
    if (/[0-9]+/.test(segment)) {
      let count = +segment
      items = pointerSchema.items = []
      pointerSchema.type = ["object", "array"]
      while (count--) items.push({***REMOVED***)
    ***REMOVED*** else {
      pointerSchema.type = "object"
    ***REMOVED***
    pointerSchema = isLast ? schema : {***REMOVED***
    properties[segment] = pointerSchema
    if (items) items.push(pointerSchema)
  ***REMOVED***
  return rootSchema
***REMOVED***

function unescapeJsonPointer(str: string): string {
  return str.replace(/~1/g, "/").replace(/~0/g, "~")
***REMOVED***

module.exports = getDef
