// ajv-keywords/src/definitions/dynamicDefaults.ts
import type {FuncKeywordDefinition, SchemaCxt***REMOVED*** from "ajv"

const sequences: Record<string, number | undefined> = {***REMOVED***

export type DynamicDefaultFunc = (args?: Record<string, any>) => () => any

const DEFAULTS: Record<string, DynamicDefaultFunc | undefined> = {
  timestamp: () => () => Date.now(),
  datetime: () => () => new Date().toISOString(),
  date: () => () => new Date().toISOString().slice(0, 10),
  time: () => () => new Date().toISOString().slice(11),
  random: () => () => Math.random(),
  randomint: (args?: {max?: number***REMOVED***) => {
    const max = args?.max ?? 2
    return () => Math.floor(Math.random() * max)
  ***REMOVED***,
  seq: (args?: {name?: string***REMOVED***) => {
    const name = args?.name ?? ""
    sequences[name] ||= 0
    return () => (sequences[name] as number)++
  ***REMOVED***,
***REMOVED***

interface PropertyDefaultSchema {
  func: string
  args: Record<string, any>
***REMOVED***

type DefaultSchema = Record<string, string | PropertyDefaultSchema | undefined>

const getDef: (() => FuncKeywordDefinition) & {
  DEFAULTS: typeof DEFAULTS
***REMOVED*** = Object.assign(_getDef, {DEFAULTS***REMOVED***)

function _getDef(): FuncKeywordDefinition {
  return {
    keyword: "dynamicDefaults",
    type: "object",
    schemaType: ["string", "object"],
    modifying: true,
    valid: true,
    compile(schema: DefaultSchema, _parentSchema, it: SchemaCxt) {
      if (!it.opts.useDefaults || it.compositeRule) return () => true
      const fs: Record<string, () => any> = {***REMOVED***
      for (const key in schema) fs[key] = getDefault(schema[key])
      const empty = it.opts.useDefaults === "empty"

      return (data: Record<string, any>) => {
        for (const prop in schema) {
          if (data[prop] === undefined || (empty && (data[prop] === null || data[prop] === ""))) {
            data[prop] = fs[prop]()
          ***REMOVED***
        ***REMOVED***
        return true
      ***REMOVED***
    ***REMOVED***,
    metaSchema: {
      type: "object",
      additionalProperties: {
        anyOf: [
          {type: "string"***REMOVED***,
          {
            type: "object",
            additionalProperties: false,
            required: ["func", "args"],
            properties: {
              func: {type: "string"***REMOVED***,
              args: {type: "object"***REMOVED***,
            ***REMOVED***,
          ***REMOVED***,
        ],
      ***REMOVED***,
    ***REMOVED***,
  ***REMOVED***
***REMOVED***

function getDefault(d: string | PropertyDefaultSchema | undefined): () => any {
  return typeof d == "object" ? getObjDefault(d) : getStrDefault(d)
***REMOVED***

function getObjDefault({func, args***REMOVED***: PropertyDefaultSchema): () => any {
  const def = DEFAULTS[func]
  assertDefined(func, def)
  return def(args)
***REMOVED***

function getStrDefault(d = ""): () => any {
  const def = DEFAULTS[d]
  assertDefined(d, def)
  return def()
***REMOVED***

function assertDefined(name: string, def?: DynamicDefaultFunc): asserts def is DynamicDefaultFunc {
  if (!def) throw new Error(`invalid "dynamicDefaults" keyword property value: ${name***REMOVED***`)
***REMOVED***

export default getDef
module.exports = getDef
