// ajv-keywords/src/definitions/patternRequired.ts
import type {CodeKeywordDefinition, KeywordCxt, KeywordErrorDefinition, ErrorObject***REMOVED*** from "ajv"
import {_, str, and***REMOVED*** from "ajv/dist/compile/codegen"
import {usePattern***REMOVED*** from "./_util"

export type PatternRequiredError = ErrorObject<"patternRequired", {missingPattern: string***REMOVED***>

const error: KeywordErrorDefinition = {
  message: ({params: {missingPattern***REMOVED******REMOVED***) =>
    str`should have property matching pattern '${missingPattern***REMOVED***'`,
  params: ({params: {missingPattern***REMOVED******REMOVED***) => _`{missingPattern: ${missingPattern***REMOVED******REMOVED***`,
***REMOVED***

export default function getDef(): CodeKeywordDefinition {
  return {
    keyword: "patternRequired",
    type: "object",
    schemaType: "array",
    error,
    code(cxt: KeywordCxt) {
      const {gen, schema, data***REMOVED*** = cxt
      if (schema.length === 0) return
      const valid = gen.let("valid", true)
      for (const pat of schema) validateProperties(pat)

      function validateProperties(pattern: string): void {
        const matched = gen.let("matched", false)

        gen.forIn("key", data, (key) => {
          gen.assign(matched, _`${usePattern(cxt, pattern)***REMOVED***.test(${key***REMOVED***)`)
          gen.if(matched, () => gen.break())
        ***REMOVED***)

        cxt.setParams({missingPattern: pattern***REMOVED***)
        gen.assign(valid, and(valid, matched))
        cxt.pass(valid)
      ***REMOVED***
    ***REMOVED***,
    metaSchema: {
      type: "array",
      items: {type: "string", format: "regex"***REMOVED***,
      uniqueItems: true,
    ***REMOVED***,
  ***REMOVED***
***REMOVED***

module.exports = getDef
