// ajv-keywords/src/definitions/regexp.ts
import type {CodeKeywordDefinition, KeywordCxt, JSONSchemaType, Name***REMOVED*** from "ajv"
import {_***REMOVED*** from "ajv/dist/compile/codegen"
import {usePattern***REMOVED*** from "./_util"

interface RegexpSchema {
  pattern: string
  flags?: string
***REMOVED***

const regexpMetaSchema: JSONSchemaType<RegexpSchema> = {
  type: "object",
  properties: {
    pattern: {type: "string"***REMOVED***,
    flags: {type: "string", nullable: true***REMOVED***,
  ***REMOVED***,
  required: ["pattern"],
  additionalProperties: false,
***REMOVED***

const metaRegexp = /^\/(.*)\/([gimuy]*)$/

export default function getDef(): CodeKeywordDefinition {
  return {
    keyword: "regexp",
    type: "string",
    schemaType: ["string", "object"],
    code(cxt: KeywordCxt) {
      const {data, schema***REMOVED*** = cxt
      const regx = getRegExp(schema)
      cxt.pass(_`${regx***REMOVED***.test(${data***REMOVED***)`)

      function getRegExp(sch: string | RegexpSchema): Name {
        if (typeof sch == "object") return usePattern(cxt, sch.pattern, sch.flags)
        const rx = metaRegexp.exec(sch)
        if (rx) return usePattern(cxt, rx[1], rx[2])
        throw new Error("cannot parse string into RegExp")
      ***REMOVED***
    ***REMOVED***,
    metaSchema: {
      anyOf: [{type: "string"***REMOVED***, regexpMetaSchema],
    ***REMOVED***,
  ***REMOVED***
***REMOVED***

module.exports = getDef
