// yocto-queue/index.js
class Node {
	/// value;
	/// next;

	constructor(value) {
		this.value = value;

		// TODO: Remove this when targeting Node.js 12.
		this.next = undefined;
	***REMOVED***
***REMOVED***

class Queue {
	// TODO: Use private class fields when targeting Node.js 12.
	// #_head;
	// #_tail;
	// #_size;

	constructor() {
		this.clear();
	***REMOVED***

	enqueue(value) {
		const node = new Node(value);

		if (this._head) {
			this._tail.next = node;
			this._tail = node;
		***REMOVED*** else {
			this._head = node;
			this._tail = node;
		***REMOVED***

		this._size++;
	***REMOVED***

	dequeue() {
		const current = this._head;
		if (!current) {
			return;
		***REMOVED***

		this._head = this._head.next;
		this._size--;
		return current.value;
	***REMOVED***

	clear() {
		this._head = undefined;
		this._tail = undefined;
		this._size = 0;
	***REMOVED***

	get size() {
		return this._size;
	***REMOVED***

	* [Symbol.iterator]() {
		let current = this._head;

		while (current) {
			yield current.value;
			current = current.next;
		***REMOVED***
	***REMOVED***
***REMOVED***

module.exports = Queue;
