// eslint/lib/rules/no-implicit-coercion.js
/**
 * @fileoverview A rule to disallow the type conversions with shorter notations.
 * @author Toru Nagashima
 */

"use strict";

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

const INDEX_OF_PATTERN = /^(?:i|lastI)ndexOf$/u;
const ALLOWABLE_OPERATORS = ["~", "!!", "+", "*"];

/**
 * Parses and normalizes an option object.
 * @param {Object***REMOVED*** options An option object to parse.
 * @returns {Object***REMOVED*** The parsed and normalized option object.
 */
function parseOptions(options) {
    return {
        boolean: "boolean" in options ? options.boolean : true,
        number: "number" in options ? options.number : true,
        string: "string" in options ? options.string : true,
        disallowTemplateShorthand: "disallowTemplateShorthand" in options ? options.disallowTemplateShorthand : false,
        allow: options.allow || []
    ***REMOVED***;
***REMOVED***

/**
 * Checks whether or not a node is a double logical negating.
 * @param {ASTNode***REMOVED*** node An UnaryExpression node to check.
 * @returns {boolean***REMOVED*** Whether or not the node is a double logical negating.
 */
function isDoubleLogicalNegating(node) {
    return (
        node.operator === "!" &&
        node.argument.type === "UnaryExpression" &&
        node.argument.operator === "!"
    );
***REMOVED***

/**
 * Checks whether or not a node is a binary negating of `.indexOf()` method calling.
 * @param {ASTNode***REMOVED*** node An UnaryExpression node to check.
 * @returns {boolean***REMOVED*** Whether or not the node is a binary negating of `.indexOf()` method calling.
 */
function isBinaryNegatingOfIndexOf(node) {
    if (node.operator !== "~") {
        return false;
    ***REMOVED***
    const callNode = astUtils.skipChainExpression(node.argument);

    return (
        callNode.type === "CallExpression" &&
        astUtils.isSpecificMemberAccess(callNode.callee, null, INDEX_OF_PATTERN)
    );
***REMOVED***

/**
 * Checks whether or not a node is a multiplying by one.
 * @param {BinaryExpression***REMOVED*** node A BinaryExpression node to check.
 * @returns {boolean***REMOVED*** Whether or not the node is a multiplying by one.
 */
function isMultiplyByOne(node) {
    return node.operator === "*" && (
        node.left.type === "Literal" && node.left.value === 1 ||
        node.right.type === "Literal" && node.right.value === 1
    );
***REMOVED***

/**
 * Checks whether the given node logically represents multiplication by a fraction of `1`.
 * For example, `a * 1` in `a * 1 / b` is technically multiplication by `1`, but the
 * whole expression can be logically interpreted as `a * (1 / b)` rather than `(a * 1) / b`.
 * @param {BinaryExpression***REMOVED*** node A BinaryExpression node to check.
 * @param {SourceCode***REMOVED*** sourceCode The source code object.
 * @returns {boolean***REMOVED*** Whether or not the node is a multiplying by a fraction of `1`.
 */
function isMultiplyByFractionOfOne(node, sourceCode) {
    return node.type === "BinaryExpression" &&
        node.operator === "*" &&
        (node.right.type === "Literal" && node.right.value === 1) &&
        node.parent.type === "BinaryExpression" &&
        node.parent.operator === "/" &&
        node.parent.left === node &&
        !astUtils.isParenthesised(sourceCode, node);
***REMOVED***

/**
 * Checks whether the result of a node is numeric or not
 * @param {ASTNode***REMOVED*** node The node to test
 * @returns {boolean***REMOVED*** true if the node is a number literal or a `Number()`, `parseInt` or `parseFloat` call
 */
function isNumeric(node) {
    return (
        node.type === "Literal" && typeof node.value === "number" ||
        node.type === "CallExpression" && (
            node.callee.name === "Number" ||
            node.callee.name === "parseInt" ||
            node.callee.name === "parseFloat"
        )
    );
***REMOVED***

/**
 * Returns the first non-numeric operand in a BinaryExpression. Designed to be
 * used from bottom to up since it walks up the BinaryExpression trees using
 * node.parent to find the result.
 * @param {BinaryExpression***REMOVED*** node The BinaryExpression node to be walked up on
 * @returns {ASTNode|null***REMOVED*** The first non-numeric item in the BinaryExpression tree or null
 */
function getNonNumericOperand(node) {
    const left = node.left,
        right = node.right;

    if (right.type !== "BinaryExpression" && !isNumeric(right)) {
        return right;
    ***REMOVED***

    if (left.type !== "BinaryExpression" && !isNumeric(left)) {
        return left;
    ***REMOVED***

    return null;
***REMOVED***

/**
 * Checks whether an expression evaluates to a string.
 * @param {ASTNode***REMOVED*** node node that represents the expression to check.
 * @returns {boolean***REMOVED*** Whether or not the expression evaluates to a string.
 */
function isStringType(node) {
    return astUtils.isStringLiteral(node) ||
        (
            node.type === "CallExpression" &&
            node.callee.type === "Identifier" &&
            node.callee.name === "String"
        );
***REMOVED***

/**
 * Checks whether a node is an empty string literal or not.
 * @param {ASTNode***REMOVED*** node The node to check.
 * @returns {boolean***REMOVED*** Whether or not the passed in node is an
 * empty string literal or not.
 */
function isEmptyString(node) {
    return astUtils.isStringLiteral(node) && (node.value === "" || (node.type === "TemplateLiteral" && node.quasis.length === 1 && node.quasis[0].value.cooked === ""));
***REMOVED***

/**
 * Checks whether or not a node is a concatenating with an empty string.
 * @param {ASTNode***REMOVED*** node A BinaryExpression node to check.
 * @returns {boolean***REMOVED*** Whether or not the node is a concatenating with an empty string.
 */
function isConcatWithEmptyString(node) {
    return node.operator === "+" && (
        (isEmptyString(node.left) && !isStringType(node.right)) ||
        (isEmptyString(node.right) && !isStringType(node.left))
    );
***REMOVED***

/**
 * Checks whether or not a node is appended with an empty string.
 * @param {ASTNode***REMOVED*** node An AssignmentExpression node to check.
 * @returns {boolean***REMOVED*** Whether or not the node is appended with an empty string.
 */
function isAppendEmptyString(node) {
    return node.operator === "+=" && isEmptyString(node.right);
***REMOVED***

/**
 * Returns the operand that is not an empty string from a flagged BinaryExpression.
 * @param {ASTNode***REMOVED*** node The flagged BinaryExpression node to check.
 * @returns {ASTNode***REMOVED*** The operand that is not an empty string from a flagged BinaryExpression.
 */
function getNonEmptyOperand(node) {
    return isEmptyString(node.left) ? node.right : node.left;
***REMOVED***

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

/** @type {import('../shared/types').Rule***REMOVED*** */
module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "Disallow shorthand type conversions",
            recommended: false,
            url: "https://eslint.org/docs/latest/rules/no-implicit-coercion"
        ***REMOVED***,

        fixable: "code",

        schema: [{
            type: "object",
            properties: {
                boolean: {
                    type: "boolean",
                    default: true
                ***REMOVED***,
                number: {
                    type: "boolean",
                    default: true
                ***REMOVED***,
                string: {
                    type: "boolean",
                    default: true
                ***REMOVED***,
                disallowTemplateShorthand: {
                    type: "boolean",
                    default: false
                ***REMOVED***,
                allow: {
                    type: "array",
                    items: {
                        enum: ALLOWABLE_OPERATORS
                    ***REMOVED***,
                    uniqueItems: true
                ***REMOVED***
            ***REMOVED***,
            additionalProperties: false
        ***REMOVED***],

        messages: {
            useRecommendation: "use `{{recommendation***REMOVED******REMOVED***` instead."
        ***REMOVED***
    ***REMOVED***,

    create(context) {
        const options = parseOptions(context.options[0] || {***REMOVED***);
        const sourceCode = context.sourceCode;

        /**
         * Reports an error and autofixes the node
         * @param {ASTNode***REMOVED*** node An ast node to report the error on.
         * @param {string***REMOVED*** recommendation The recommended code for the issue
         * @param {bool***REMOVED*** shouldFix Whether this report should fix the node
         * @returns {void***REMOVED***
         */
        function report(node, recommendation, shouldFix) {
            context.report({
                node,
                messageId: "useRecommendation",
                data: {
                    recommendation
                ***REMOVED***,
                fix(fixer) {
                    if (!shouldFix) {
                        return null;
                    ***REMOVED***

                    const tokenBefore = sourceCode.getTokenBefore(node);

                    if (
                        tokenBefore &&
                        tokenBefore.range[1] === node.range[0] &&
                        !astUtils.canTokensBeAdjacent(tokenBefore, recommendation)
                    ) {
                        return fixer.replaceText(node, ` ${recommendation***REMOVED***`);
                    ***REMOVED***
                    return fixer.replaceText(node, recommendation);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        return {
            UnaryExpression(node) {
                let operatorAllowed;

                // !!foo
                operatorAllowed = options.allow.includes("!!");
                if (!operatorAllowed && options.boolean && isDoubleLogicalNegating(node)) {
                    const recommendation = `Boolean(${sourceCode.getText(node.argument.argument)***REMOVED***)`;

                    report(node, recommendation, true);
                ***REMOVED***

                // ~foo.indexOf(bar)
                operatorAllowed = options.allow.includes("~");
                if (!operatorAllowed && options.boolean && isBinaryNegatingOfIndexOf(node)) {

                    // `foo?.indexOf(bar) !== -1` will be true (== found) if the `foo` is nullish. So use `>= 0` in that case.
                    const comparison = node.argument.type === "ChainExpression" ? ">= 0" : "!== -1";
                    const recommendation = `${sourceCode.getText(node.argument)***REMOVED*** ${comparison***REMOVED***`;

                    report(node, recommendation, false);
                ***REMOVED***

                // +foo
                operatorAllowed = options.allow.includes("+");
                if (!operatorAllowed && options.number && node.operator === "+" && !isNumeric(node.argument)) {
                    const recommendation = `Number(${sourceCode.getText(node.argument)***REMOVED***)`;

                    report(node, recommendation, true);
                ***REMOVED***
            ***REMOVED***,

            // Use `:exit` to prevent double reporting
            "BinaryExpression:exit"(node) {
                let operatorAllowed;

                // 1 * foo
                operatorAllowed = options.allow.includes("*");
                const nonNumericOperand = !operatorAllowed && options.number && isMultiplyByOne(node) && !isMultiplyByFractionOfOne(node, sourceCode) &&
                    getNonNumericOperand(node);

                if (nonNumericOperand) {
                    const recommendation = `Number(${sourceCode.getText(nonNumericOperand)***REMOVED***)`;

                    report(node, recommendation, true);
                ***REMOVED***

                // "" + foo
                operatorAllowed = options.allow.includes("+");
                if (!operatorAllowed && options.string && isConcatWithEmptyString(node)) {
                    const recommendation = `String(${sourceCode.getText(getNonEmptyOperand(node))***REMOVED***)`;

                    report(node, recommendation, true);
                ***REMOVED***
            ***REMOVED***,

            AssignmentExpression(node) {

                // foo += ""
                const operatorAllowed = options.allow.includes("+");

                if (!operatorAllowed && options.string && isAppendEmptyString(node)) {
                    const code = sourceCode.getText(getNonEmptyOperand(node));
                    const recommendation = `${code***REMOVED*** = String(${code***REMOVED***)`;

                    report(node, recommendation, true);
                ***REMOVED***
            ***REMOVED***,

            TemplateLiteral(node) {
                if (!options.disallowTemplateShorthand) {
                    return;
                ***REMOVED***

                // tag`${foo***REMOVED***`
                if (node.parent.type === "TaggedTemplateExpression") {
                    return;
                ***REMOVED***

                // `` or `${foo***REMOVED***${bar***REMOVED***`
                if (node.expressions.length !== 1) {
                    return;
                ***REMOVED***


                //  `prefix${foo***REMOVED***`
                if (node.quasis[0].value.cooked !== "") {
                    return;
                ***REMOVED***

                //  `${foo***REMOVED***postfix`
                if (node.quasis[1].value.cooked !== "") {
                    return;
                ***REMOVED***

                // if the expression is already a string, then this isn't a coercion
                if (isStringType(node.expressions[0])) {
                    return;
                ***REMOVED***

                const code = sourceCode.getText(node.expressions[0]);
                const recommendation = `String(${code***REMOVED***)`;

                report(node, recommendation, true);
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
