// eslint/lib/rules/object-curly-spacing.js
/**
 * @fileoverview Disallows or enforces spaces inside of object literals.
 * @author Jamund Ferguson
 * @deprecated in ESLint v8.53.0
 */
"use strict";

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

/** @type {import('../shared/types').Rule***REMOVED*** */
module.exports = {
    meta: {
        deprecated: true,
        replacedBy: [],
        type: "layout",

        docs: {
            description: "Enforce consistent spacing inside braces",
            recommended: false,
            url: "https://eslint.org/docs/latest/rules/object-curly-spacing"
        ***REMOVED***,

        fixable: "whitespace",

        schema: [
            {
                enum: ["always", "never"]
            ***REMOVED***,
            {
                type: "object",
                properties: {
                    arraysInObjects: {
                        type: "boolean"
                    ***REMOVED***,
                    objectsInObjects: {
                        type: "boolean"
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],

        messages: {
            requireSpaceBefore: "A space is required before '{{token***REMOVED******REMOVED***'.",
            requireSpaceAfter: "A space is required after '{{token***REMOVED******REMOVED***'.",
            unexpectedSpaceBefore: "There should be no space before '{{token***REMOVED******REMOVED***'.",
            unexpectedSpaceAfter: "There should be no space after '{{token***REMOVED******REMOVED***'."
        ***REMOVED***
    ***REMOVED***,

    create(context) {
        const spaced = context.options[0] === "always",
            sourceCode = context.sourceCode;

        /**
         * Determines whether an option is set, relative to the spacing option.
         * If spaced is "always", then check whether option is set to false.
         * If spaced is "never", then check whether option is set to true.
         * @param {Object***REMOVED*** option The option to exclude.
         * @returns {boolean***REMOVED*** Whether or not the property is excluded.
         */
        function isOptionSet(option) {
            return context.options[1] ? context.options[1][option] === !spaced : false;
        ***REMOVED***

        const options = {
            spaced,
            arraysInObjectsException: isOptionSet("arraysInObjects"),
            objectsInObjectsException: isOptionSet("objectsInObjects")
        ***REMOVED***;

        //--------------------------------------------------------------------------
        // Helpers
        //--------------------------------------------------------------------------

        /**
         * Reports that there shouldn't be a space after the first token
         * @param {ASTNode***REMOVED*** node The node to report in the event of an error.
         * @param {Token***REMOVED*** token The token to use for the report.
         * @returns {void***REMOVED***
         */
        function reportNoBeginningSpace(node, token) {
            const nextToken = context.sourceCode.getTokenAfter(token, { includeComments: true ***REMOVED***);

            context.report({
                node,
                loc: { start: token.loc.end, end: nextToken.loc.start ***REMOVED***,
                messageId: "unexpectedSpaceAfter",
                data: {
                    token: token.value
                ***REMOVED***,
                fix(fixer) {
                    return fixer.removeRange([token.range[1], nextToken.range[0]]);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Reports that there shouldn't be a space before the last token
         * @param {ASTNode***REMOVED*** node The node to report in the event of an error.
         * @param {Token***REMOVED*** token The token to use for the report.
         * @returns {void***REMOVED***
         */
        function reportNoEndingSpace(node, token) {
            const previousToken = context.sourceCode.getTokenBefore(token, { includeComments: true ***REMOVED***);

            context.report({
                node,
                loc: { start: previousToken.loc.end, end: token.loc.start ***REMOVED***,
                messageId: "unexpectedSpaceBefore",
                data: {
                    token: token.value
                ***REMOVED***,
                fix(fixer) {
                    return fixer.removeRange([previousToken.range[1], token.range[0]]);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Reports that there should be a space after the first token
         * @param {ASTNode***REMOVED*** node The node to report in the event of an error.
         * @param {Token***REMOVED*** token The token to use for the report.
         * @returns {void***REMOVED***
         */
        function reportRequiredBeginningSpace(node, token) {
            context.report({
                node,
                loc: token.loc,
                messageId: "requireSpaceAfter",
                data: {
                    token: token.value
                ***REMOVED***,
                fix(fixer) {
                    return fixer.insertTextAfter(token, " ");
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Reports that there should be a space before the last token
         * @param {ASTNode***REMOVED*** node The node to report in the event of an error.
         * @param {Token***REMOVED*** token The token to use for the report.
         * @returns {void***REMOVED***
         */
        function reportRequiredEndingSpace(node, token) {
            context.report({
                node,
                loc: token.loc,
                messageId: "requireSpaceBefore",
                data: {
                    token: token.value
                ***REMOVED***,
                fix(fixer) {
                    return fixer.insertTextBefore(token, " ");
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Determines if spacing in curly braces is valid.
         * @param {ASTNode***REMOVED*** node The AST node to check.
         * @param {Token***REMOVED*** first The first token to check (should be the opening brace)
         * @param {Token***REMOVED*** second The second token to check (should be first after the opening brace)
         * @param {Token***REMOVED*** penultimate The penultimate token to check (should be last before closing brace)
         * @param {Token***REMOVED*** last The last token to check (should be closing brace)
         * @returns {void***REMOVED***
         */
        function validateBraceSpacing(node, first, second, penultimate, last) {
            if (astUtils.isTokenOnSameLine(first, second)) {
                const firstSpaced = sourceCode.isSpaceBetweenTokens(first, second);

                if (options.spaced && !firstSpaced) {
                    reportRequiredBeginningSpace(node, first);
                ***REMOVED***
                if (!options.spaced && firstSpaced && second.type !== "Line") {
                    reportNoBeginningSpace(node, first);
                ***REMOVED***
            ***REMOVED***

            if (astUtils.isTokenOnSameLine(penultimate, last)) {
                const shouldCheckPenultimate = (
                    options.arraysInObjectsException && astUtils.isClosingBracketToken(penultimate) ||
                    options.objectsInObjectsException && astUtils.isClosingBraceToken(penultimate)
                );
                const penultimateType = shouldCheckPenultimate && sourceCode.getNodeByRangeIndex(penultimate.range[0]).type;

                const closingCurlyBraceMustBeSpaced = (
                    options.arraysInObjectsException && penultimateType === "ArrayExpression" ||
                    options.objectsInObjectsException && (penultimateType === "ObjectExpression" || penultimateType === "ObjectPattern")
                ) ? !options.spaced : options.spaced;

                const lastSpaced = sourceCode.isSpaceBetweenTokens(penultimate, last);

                if (closingCurlyBraceMustBeSpaced && !lastSpaced) {
                    reportRequiredEndingSpace(node, last);
                ***REMOVED***
                if (!closingCurlyBraceMustBeSpaced && lastSpaced) {
                    reportNoEndingSpace(node, last);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        /**
         * Gets '***REMOVED***' token of an object node.
         *
         * Because the last token of object patterns might be a type annotation,
         * this traverses tokens preceded by the last property, then returns the
         * first '***REMOVED***' token.
         * @param {ASTNode***REMOVED*** node The node to get. This node is an
         *      ObjectExpression or an ObjectPattern. And this node has one or
         *      more properties.
         * @returns {Token***REMOVED*** '***REMOVED***' token.
         */
        function getClosingBraceOfObject(node) {
            const lastProperty = node.properties[node.properties.length - 1];

            return sourceCode.getTokenAfter(lastProperty, astUtils.isClosingBraceToken);
        ***REMOVED***

        /**
         * Reports a given object node if spacing in curly braces is invalid.
         * @param {ASTNode***REMOVED*** node An ObjectExpression or ObjectPattern node to check.
         * @returns {void***REMOVED***
         */
        function checkForObject(node) {
            if (node.properties.length === 0) {
                return;
            ***REMOVED***

            const first = sourceCode.getFirstToken(node),
                last = getClosingBraceOfObject(node),
                second = sourceCode.getTokenAfter(first, { includeComments: true ***REMOVED***),
                penultimate = sourceCode.getTokenBefore(last, { includeComments: true ***REMOVED***);

            validateBraceSpacing(node, first, second, penultimate, last);
        ***REMOVED***

        /**
         * Reports a given import node if spacing in curly braces is invalid.
         * @param {ASTNode***REMOVED*** node An ImportDeclaration node to check.
         * @returns {void***REMOVED***
         */
        function checkForImport(node) {
            if (node.specifiers.length === 0) {
                return;
            ***REMOVED***

            let firstSpecifier = node.specifiers[0];
            const lastSpecifier = node.specifiers[node.specifiers.length - 1];

            if (lastSpecifier.type !== "ImportSpecifier") {
                return;
            ***REMOVED***
            if (firstSpecifier.type !== "ImportSpecifier") {
                firstSpecifier = node.specifiers[1];
            ***REMOVED***

            const first = sourceCode.getTokenBefore(firstSpecifier),
                last = sourceCode.getTokenAfter(lastSpecifier, astUtils.isNotCommaToken),
                second = sourceCode.getTokenAfter(first, { includeComments: true ***REMOVED***),
                penultimate = sourceCode.getTokenBefore(last, { includeComments: true ***REMOVED***);

            validateBraceSpacing(node, first, second, penultimate, last);
        ***REMOVED***

        /**
         * Reports a given export node if spacing in curly braces is invalid.
         * @param {ASTNode***REMOVED*** node An ExportNamedDeclaration node to check.
         * @returns {void***REMOVED***
         */
        function checkForExport(node) {
            if (node.specifiers.length === 0) {
                return;
            ***REMOVED***

            const firstSpecifier = node.specifiers[0],
                lastSpecifier = node.specifiers[node.specifiers.length - 1],
                first = sourceCode.getTokenBefore(firstSpecifier),
                last = sourceCode.getTokenAfter(lastSpecifier, astUtils.isNotCommaToken),
                second = sourceCode.getTokenAfter(first, { includeComments: true ***REMOVED***),
                penultimate = sourceCode.getTokenBefore(last, { includeComments: true ***REMOVED***);

            validateBraceSpacing(node, first, second, penultimate, last);
        ***REMOVED***

        //--------------------------------------------------------------------------
        // Public
        //--------------------------------------------------------------------------

        return {

            // var {x***REMOVED*** = y;
            ObjectPattern: checkForObject,

            // var y = {x: 'y'***REMOVED***
            ObjectExpression: checkForObject,

            // import {y***REMOVED*** from 'x';
            ImportDeclaration: checkForImport,

            // export {name***REMOVED*** from 'yo';
            ExportNamedDeclaration: checkForExport
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
