// eslint/lib/rules/quote-props.js
/**
 * @fileoverview Rule to flag non-quoted property names in object literals.
 * @author Mathias Bynens <http://mathiasbynens.be/>
 * @deprecated in ESLint v8.53.0
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const espree = require("espree");
const astUtils = require("./utils/ast-utils");
const keywords = require("./utils/keywords");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

/** @type {import('../shared/types').Rule***REMOVED*** */
module.exports = {
    meta: {
        deprecated: true,
        replacedBy: [],
        type: "suggestion",

        docs: {
            description: "Require quotes around object literal property names",
            recommended: false,
            url: "https://eslint.org/docs/latest/rules/quote-props"
        ***REMOVED***,

        schema: {
            anyOf: [
                {
                    type: "array",
                    items: [
                        {
                            enum: ["always", "as-needed", "consistent", "consistent-as-needed"]
                        ***REMOVED***
                    ],
                    minItems: 0,
                    maxItems: 1
                ***REMOVED***,
                {
                    type: "array",
                    items: [
                        {
                            enum: ["always", "as-needed", "consistent", "consistent-as-needed"]
                        ***REMOVED***,
                        {
                            type: "object",
                            properties: {
                                keywords: {
                                    type: "boolean"
                                ***REMOVED***,
                                unnecessary: {
                                    type: "boolean"
                                ***REMOVED***,
                                numbers: {
                                    type: "boolean"
                                ***REMOVED***
                            ***REMOVED***,
                            additionalProperties: false
                        ***REMOVED***
                    ],
                    minItems: 0,
                    maxItems: 2
                ***REMOVED***
            ]
        ***REMOVED***,

        fixable: "code",
        messages: {
            requireQuotesDueToReservedWord: "Properties should be quoted as '{{property***REMOVED******REMOVED***' is a reserved word.",
            inconsistentlyQuotedProperty: "Inconsistently quoted property '{{key***REMOVED******REMOVED***' found.",
            unnecessarilyQuotedProperty: "Unnecessarily quoted property '{{property***REMOVED******REMOVED***' found.",
            unquotedReservedProperty: "Unquoted reserved word '{{property***REMOVED******REMOVED***' used as key.",
            unquotedNumericProperty: "Unquoted number literal '{{property***REMOVED******REMOVED***' used as key.",
            unquotedPropertyFound: "Unquoted property '{{property***REMOVED******REMOVED***' found.",
            redundantQuoting: "Properties shouldn't be quoted as all quotes are redundant."
        ***REMOVED***
    ***REMOVED***,

    create(context) {

        const MODE = context.options[0],
            KEYWORDS = context.options[1] && context.options[1].keywords,
            CHECK_UNNECESSARY = !context.options[1] || context.options[1].unnecessary !== false,
            NUMBERS = context.options[1] && context.options[1].numbers,

            sourceCode = context.sourceCode;


        /**
         * Checks whether a certain string constitutes an ES3 token
         * @param {string***REMOVED*** tokenStr The string to be checked.
         * @returns {boolean***REMOVED*** `true` if it is an ES3 token.
         */
        function isKeyword(tokenStr) {
            return keywords.includes(tokenStr);
        ***REMOVED***

        /**
         * Checks if an espree-tokenized key has redundant quotes (i.e. whether quotes are unnecessary)
         * @param {string***REMOVED*** rawKey The raw key value from the source
         * @param {espreeTokens***REMOVED*** tokens The espree-tokenized node key
         * @param {boolean***REMOVED*** [skipNumberLiterals=false] Indicates whether number literals should be checked
         * @returns {boolean***REMOVED*** Whether or not a key has redundant quotes.
         * @private
         */
        function areQuotesRedundant(rawKey, tokens, skipNumberLiterals) {
            return tokens.length === 1 && tokens[0].start === 0 && tokens[0].end === rawKey.length &&
                (["Identifier", "Keyword", "Null", "Boolean"].includes(tokens[0].type) ||
                (tokens[0].type === "Numeric" && !skipNumberLiterals && String(+tokens[0].value) === tokens[0].value));
        ***REMOVED***

        /**
         * Returns a string representation of a property node with quotes removed
         * @param {ASTNode***REMOVED*** key Key AST Node, which may or may not be quoted
         * @returns {string***REMOVED*** A replacement string for this property
         */
        function getUnquotedKey(key) {
            return key.type === "Identifier" ? key.name : key.value;
        ***REMOVED***

        /**
         * Returns a string representation of a property node with quotes added
         * @param {ASTNode***REMOVED*** key Key AST Node, which may or may not be quoted
         * @returns {string***REMOVED*** A replacement string for this property
         */
        function getQuotedKey(key) {
            if (key.type === "Literal" && typeof key.value === "string") {

                // If the key is already a string literal, don't replace the quotes with double quotes.
                return sourceCode.getText(key);
            ***REMOVED***

            // Otherwise, the key is either an identifier or a number literal.
            return `"${key.type === "Identifier" ? key.name : key.value***REMOVED***"`;
        ***REMOVED***

        /**
         * Ensures that a property's key is quoted only when necessary
         * @param {ASTNode***REMOVED*** node Property AST node
         * @returns {void***REMOVED***
         */
        function checkUnnecessaryQuotes(node) {
            const key = node.key;

            if (node.method || node.computed || node.shorthand) {
                return;
            ***REMOVED***

            if (key.type === "Literal" && typeof key.value === "string") {
                let tokens;

                try {
                    tokens = espree.tokenize(key.value);
                ***REMOVED*** catch {
                    return;
                ***REMOVED***

                if (tokens.length !== 1) {
                    return;
                ***REMOVED***

                const isKeywordToken = isKeyword(tokens[0].value);

                if (isKeywordToken && KEYWORDS) {
                    return;
                ***REMOVED***

                if (CHECK_UNNECESSARY && areQuotesRedundant(key.value, tokens, NUMBERS)) {
                    context.report({
                        node,
                        messageId: "unnecessarilyQuotedProperty",
                        data: { property: key.value ***REMOVED***,
                        fix: fixer => fixer.replaceText(key, getUnquotedKey(key))
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED*** else if (KEYWORDS && key.type === "Identifier" && isKeyword(key.name)) {
                context.report({
                    node,
                    messageId: "unquotedReservedProperty",
                    data: { property: key.name ***REMOVED***,
                    fix: fixer => fixer.replaceText(key, getQuotedKey(key))
                ***REMOVED***);
            ***REMOVED*** else if (NUMBERS && key.type === "Literal" && astUtils.isNumericLiteral(key)) {
                context.report({
                    node,
                    messageId: "unquotedNumericProperty",
                    data: { property: key.value ***REMOVED***,
                    fix: fixer => fixer.replaceText(key, getQuotedKey(key))
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        /**
         * Ensures that a property's key is quoted
         * @param {ASTNode***REMOVED*** node Property AST node
         * @returns {void***REMOVED***
         */
        function checkOmittedQuotes(node) {
            const key = node.key;

            if (!node.method && !node.computed && !node.shorthand && !(key.type === "Literal" && typeof key.value === "string")) {
                context.report({
                    node,
                    messageId: "unquotedPropertyFound",
                    data: { property: key.name || key.value ***REMOVED***,
                    fix: fixer => fixer.replaceText(key, getQuotedKey(key))
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        /**
         * Ensures that an object's keys are consistently quoted, optionally checks for redundancy of quotes
         * @param {ASTNode***REMOVED*** node Property AST node
         * @param {boolean***REMOVED*** checkQuotesRedundancy Whether to check quotes' redundancy
         * @returns {void***REMOVED***
         */
        function checkConsistency(node, checkQuotesRedundancy) {
            const quotedProps = [],
                unquotedProps = [];
            let keywordKeyName = null,
                necessaryQuotes = false;

            node.properties.forEach(property => {
                const key = property.key;

                if (!key || property.method || property.computed || property.shorthand) {
                    return;
                ***REMOVED***

                if (key.type === "Literal" && typeof key.value === "string") {

                    quotedProps.push(property);

                    if (checkQuotesRedundancy) {
                        let tokens;

                        try {
                            tokens = espree.tokenize(key.value);
                        ***REMOVED*** catch {
                            necessaryQuotes = true;
                            return;
                        ***REMOVED***

                        necessaryQuotes = necessaryQuotes || !areQuotesRedundant(key.value, tokens) || KEYWORDS && isKeyword(tokens[0].value);
                    ***REMOVED***
                ***REMOVED*** else if (KEYWORDS && checkQuotesRedundancy && key.type === "Identifier" && isKeyword(key.name)) {
                    unquotedProps.push(property);
                    necessaryQuotes = true;
                    keywordKeyName = key.name;
                ***REMOVED*** else {
                    unquotedProps.push(property);
                ***REMOVED***
            ***REMOVED***);

            if (checkQuotesRedundancy && quotedProps.length && !necessaryQuotes) {
                quotedProps.forEach(property => {
                    context.report({
                        node: property,
                        messageId: "redundantQuoting",
                        fix: fixer => fixer.replaceText(property.key, getUnquotedKey(property.key))
                    ***REMOVED***);
                ***REMOVED***);
            ***REMOVED*** else if (unquotedProps.length && keywordKeyName) {
                unquotedProps.forEach(property => {
                    context.report({
                        node: property,
                        messageId: "requireQuotesDueToReservedWord",
                        data: { property: keywordKeyName ***REMOVED***,
                        fix: fixer => fixer.replaceText(property.key, getQuotedKey(property.key))
                    ***REMOVED***);
                ***REMOVED***);
            ***REMOVED*** else if (quotedProps.length && unquotedProps.length) {
                unquotedProps.forEach(property => {
                    context.report({
                        node: property,
                        messageId: "inconsistentlyQuotedProperty",
                        data: { key: property.key.name || property.key.value ***REMOVED***,
                        fix: fixer => fixer.replaceText(property.key, getQuotedKey(property.key))
                    ***REMOVED***);
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        return {
            Property(node) {
                if (MODE === "always" || !MODE) {
                    checkOmittedQuotes(node);
                ***REMOVED***
                if (MODE === "as-needed") {
                    checkUnnecessaryQuotes(node);
                ***REMOVED***
            ***REMOVED***,
            ObjectExpression(node) {
                if (MODE === "consistent") {
                    checkConsistency(node, false);
                ***REMOVED***
                if (MODE === "consistent-as-needed") {
                    checkConsistency(node, true);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
