// @webassemblyjs/ast/lib/definitions.js
"use strict";

var definitions = {***REMOVED***;

function defineType(typeName, metadata) {
  definitions[typeName] = metadata;
***REMOVED***

defineType("Module", {
  spec: {
    wasm: "https://webassembly.github.io/spec/core/binary/modules.html#binary-module",
    wat: "https://webassembly.github.io/spec/core/text/modules.html#text-module"
  ***REMOVED***,
  doc: "A module consists of a sequence of sections (termed fields in the text format).",
  unionType: ["Node"],
  fields: {
    id: {
      maybe: true,
      type: "string"
    ***REMOVED***,
    fields: {
      array: true,
      type: "Node"
    ***REMOVED***,
    metadata: {
      optional: true,
      type: "ModuleMetadata"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("ModuleMetadata", {
  unionType: ["Node"],
  fields: {
    sections: {
      array: true,
      type: "SectionMetadata"
    ***REMOVED***,
    functionNames: {
      optional: true,
      array: true,
      type: "FunctionNameMetadata"
    ***REMOVED***,
    localNames: {
      optional: true,
      array: true,
      type: "ModuleMetadata"
    ***REMOVED***,
    producers: {
      optional: true,
      array: true,
      type: "ProducersSectionMetadata"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("ModuleNameMetadata", {
  unionType: ["Node"],
  fields: {
    value: {
      type: "string"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("FunctionNameMetadata", {
  unionType: ["Node"],
  fields: {
    value: {
      type: "string"
    ***REMOVED***,
    index: {
      type: "number"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("LocalNameMetadata", {
  unionType: ["Node"],
  fields: {
    value: {
      type: "string"
    ***REMOVED***,
    localIndex: {
      type: "number"
    ***REMOVED***,
    functionIndex: {
      type: "number"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("BinaryModule", {
  unionType: ["Node"],
  fields: {
    id: {
      maybe: true,
      type: "string"
    ***REMOVED***,
    blob: {
      array: true,
      type: "string"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("QuoteModule", {
  unionType: ["Node"],
  fields: {
    id: {
      maybe: true,
      type: "string"
    ***REMOVED***,
    string: {
      array: true,
      type: "string"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("SectionMetadata", {
  unionType: ["Node"],
  fields: {
    section: {
      type: "SectionName"
    ***REMOVED***,
    startOffset: {
      type: "number"
    ***REMOVED***,
    size: {
      type: "NumberLiteral"
    ***REMOVED***,
    vectorOfSize: {
      comment: "Size of the vector in the section (if any)",
      type: "NumberLiteral"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("ProducersSectionMetadata", {
  unionType: ["Node"],
  fields: {
    producers: {
      array: true,
      type: "ProducerMetadata"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("ProducerMetadata", {
  unionType: ["Node"],
  fields: {
    language: {
      type: "ProducerMetadataVersionedName",
      array: true
    ***REMOVED***,
    processedBy: {
      type: "ProducerMetadataVersionedName",
      array: true
    ***REMOVED***,
    sdk: {
      type: "ProducerMetadataVersionedName",
      array: true
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("ProducerMetadataVersionedName", {
  unionType: ["Node"],
  fields: {
    name: {
      type: "string"
    ***REMOVED***,
    version: {
      type: "string"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
/*
Instructions
*/

defineType("LoopInstruction", {
  unionType: ["Node", "Block", "Instruction"],
  fields: {
    id: {
      constant: true,
      type: "string",
      value: "loop"
    ***REMOVED***,
    label: {
      maybe: true,
      type: "Identifier"
    ***REMOVED***,
    resulttype: {
      maybe: true,
      type: "Valtype"
    ***REMOVED***,
    instr: {
      array: true,
      type: "Instruction"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("Instr", {
  unionType: ["Node", "Expression", "Instruction"],
  fields: {
    id: {
      type: "string"
    ***REMOVED***,
    object: {
      optional: true,
      type: "Valtype"
    ***REMOVED***,
    args: {
      array: true,
      type: "Expression"
    ***REMOVED***,
    namedArgs: {
      optional: true,
      type: "Object"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("IfInstruction", {
  unionType: ["Node", "Instruction"],
  fields: {
    id: {
      constant: true,
      type: "string",
      value: "if"
    ***REMOVED***,
    testLabel: {
      comment: "only for WAST",
      type: "Identifier"
    ***REMOVED***,
    test: {
      array: true,
      type: "Instruction"
    ***REMOVED***,
    result: {
      maybe: true,
      type: "Valtype"
    ***REMOVED***,
    consequent: {
      array: true,
      type: "Instruction"
    ***REMOVED***,
    alternate: {
      array: true,
      type: "Instruction"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
/* 
Concrete value types
*/

defineType("StringLiteral", {
  unionType: ["Node", "Expression"],
  fields: {
    value: {
      type: "string"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("NumberLiteral", {
  unionType: ["Node", "NumericLiteral", "Expression"],
  fields: {
    value: {
      type: "number"
    ***REMOVED***,
    raw: {
      type: "string"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("LongNumberLiteral", {
  unionType: ["Node", "NumericLiteral", "Expression"],
  fields: {
    value: {
      type: "LongNumber"
    ***REMOVED***,
    raw: {
      type: "string"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("FloatLiteral", {
  unionType: ["Node", "NumericLiteral", "Expression"],
  fields: {
    value: {
      type: "number"
    ***REMOVED***,
    nan: {
      optional: true,
      type: "boolean"
    ***REMOVED***,
    inf: {
      optional: true,
      type: "boolean"
    ***REMOVED***,
    raw: {
      type: "string"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("Elem", {
  unionType: ["Node"],
  fields: {
    table: {
      type: "Index"
    ***REMOVED***,
    offset: {
      array: true,
      type: "Instruction"
    ***REMOVED***,
    funcs: {
      array: true,
      type: "Index"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("IndexInFuncSection", {
  unionType: ["Node"],
  fields: {
    index: {
      type: "Index"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("ValtypeLiteral", {
  unionType: ["Node", "Expression"],
  fields: {
    name: {
      type: "Valtype"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("TypeInstruction", {
  unionType: ["Node", "Instruction"],
  fields: {
    id: {
      maybe: true,
      type: "Index"
    ***REMOVED***,
    functype: {
      type: "Signature"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("Start", {
  unionType: ["Node"],
  fields: {
    index: {
      type: "Index"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("GlobalType", {
  unionType: ["Node", "ImportDescr"],
  fields: {
    valtype: {
      type: "Valtype"
    ***REMOVED***,
    mutability: {
      type: "Mutability"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("LeadingComment", {
  unionType: ["Node"],
  fields: {
    value: {
      type: "string"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("BlockComment", {
  unionType: ["Node"],
  fields: {
    value: {
      type: "string"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("Data", {
  unionType: ["Node"],
  fields: {
    memoryIndex: {
      type: "Memidx"
    ***REMOVED***,
    offset: {
      type: "Instruction"
    ***REMOVED***,
    init: {
      type: "ByteArray"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("Global", {
  unionType: ["Node"],
  fields: {
    globalType: {
      type: "GlobalType"
    ***REMOVED***,
    init: {
      array: true,
      type: "Instruction"
    ***REMOVED***,
    name: {
      maybe: true,
      type: "Identifier"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("Table", {
  unionType: ["Node", "ImportDescr"],
  fields: {
    elementType: {
      type: "TableElementType"
    ***REMOVED***,
    limits: {
      assertNodeType: true,
      type: "Limit"
    ***REMOVED***,
    name: {
      maybe: true,
      type: "Identifier"
    ***REMOVED***,
    elements: {
      array: true,
      optional: true,
      type: "Index"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("Memory", {
  unionType: ["Node", "ImportDescr"],
  fields: {
    limits: {
      type: "Limit"
    ***REMOVED***,
    id: {
      maybe: true,
      type: "Index"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("FuncImportDescr", {
  unionType: ["Node", "ImportDescr"],
  fields: {
    id: {
      type: "Identifier"
    ***REMOVED***,
    signature: {
      type: "Signature"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("ModuleImport", {
  unionType: ["Node"],
  fields: {
    module: {
      type: "string"
    ***REMOVED***,
    name: {
      type: "string"
    ***REMOVED***,
    descr: {
      type: "ImportDescr"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("ModuleExportDescr", {
  unionType: ["Node"],
  fields: {
    exportType: {
      type: "ExportDescrType"
    ***REMOVED***,
    id: {
      type: "Index"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("ModuleExport", {
  unionType: ["Node"],
  fields: {
    name: {
      type: "string"
    ***REMOVED***,
    descr: {
      type: "ModuleExportDescr"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("Limit", {
  unionType: ["Node"],
  fields: {
    min: {
      type: "number"
    ***REMOVED***,
    max: {
      optional: true,
      type: "number"
    ***REMOVED***,
    // Threads proposal, shared memory
    shared: {
      optional: true,
      type: "boolean"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("Signature", {
  unionType: ["Node"],
  fields: {
    params: {
      array: true,
      type: "FuncParam"
    ***REMOVED***,
    results: {
      array: true,
      type: "Valtype"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("Program", {
  unionType: ["Node"],
  fields: {
    body: {
      array: true,
      type: "Node"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("Identifier", {
  unionType: ["Node", "Expression"],
  fields: {
    value: {
      type: "string"
    ***REMOVED***,
    raw: {
      optional: true,
      type: "string"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("BlockInstruction", {
  unionType: ["Node", "Block", "Instruction"],
  fields: {
    id: {
      constant: true,
      type: "string",
      value: "block"
    ***REMOVED***,
    label: {
      maybe: true,
      type: "Identifier"
    ***REMOVED***,
    instr: {
      array: true,
      type: "Instruction"
    ***REMOVED***,
    result: {
      maybe: true,
      type: "Valtype"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("CallInstruction", {
  unionType: ["Node", "Instruction"],
  fields: {
    id: {
      constant: true,
      type: "string",
      value: "call"
    ***REMOVED***,
    index: {
      type: "Index"
    ***REMOVED***,
    instrArgs: {
      array: true,
      optional: true,
      type: "Expression"
    ***REMOVED***,
    numeric: {
      type: "Index",
      optional: true
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("CallIndirectInstruction", {
  unionType: ["Node", "Instruction"],
  fields: {
    id: {
      constant: true,
      type: "string",
      value: "call_indirect"
    ***REMOVED***,
    signature: {
      type: "SignatureOrTypeRef"
    ***REMOVED***,
    intrs: {
      array: true,
      optional: true,
      type: "Expression"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("ByteArray", {
  unionType: ["Node"],
  fields: {
    values: {
      array: true,
      type: "Byte"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("Func", {
  unionType: ["Node", "Block"],
  fields: {
    name: {
      maybe: true,
      type: "Index"
    ***REMOVED***,
    signature: {
      type: "SignatureOrTypeRef"
    ***REMOVED***,
    body: {
      array: true,
      type: "Instruction"
    ***REMOVED***,
    isExternal: {
      comment: "means that it has been imported from the outside js",
      optional: true,
      type: "boolean"
    ***REMOVED***,
    metadata: {
      optional: true,
      type: "FuncMetadata"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
/**
 * Intrinsics
 */

defineType("InternalBrUnless", {
  unionType: ["Node", "Intrinsic"],
  fields: {
    target: {
      type: "number"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("InternalGoto", {
  unionType: ["Node", "Intrinsic"],
  fields: {
    target: {
      type: "number"
    ***REMOVED***
  ***REMOVED***
***REMOVED***);
defineType("InternalCallExtern", {
  unionType: ["Node", "Intrinsic"],
  fields: {
    target: {
      type: "number"
    ***REMOVED***
  ***REMOVED***
***REMOVED***); // function bodies are terminated by an `end` instruction but are missing a
// return instruction
//
// Since we can't inject a new instruction we are injecting a new instruction.

defineType("InternalEndAndReturn", {
  unionType: ["Node", "Intrinsic"],
  fields: {***REMOVED***
***REMOVED***);
module.exports = definitions;
