// @humanwhocodes/config-array/api.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true ***REMOVED***);

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; ***REMOVED***

var path = _interopDefault(require('path'));
var minimatch = _interopDefault(require('minimatch'));
var createDebug = _interopDefault(require('debug'));
var objectSchema = require('@humanwhocodes/object-schema');

/**
 * @fileoverview ConfigSchema
 * @author Nicholas C. Zakas
 */

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Assets that a given value is an array.
 * @param {****REMOVED*** value The value to check.
 * @returns {void***REMOVED***
 * @throws {TypeError***REMOVED*** When the value is not an array. 
 */
function assertIsArray(value) {
	if (!Array.isArray(value)) {
		throw new TypeError('Expected value to be an array.');
	***REMOVED***
***REMOVED***

/**
 * Assets that a given value is an array containing only strings and functions.
 * @param {****REMOVED*** value The value to check.
 * @returns {void***REMOVED***
 * @throws {TypeError***REMOVED*** When the value is not an array of strings and functions.
 */
function assertIsArrayOfStringsAndFunctions(value, name) {
	assertIsArray(value);

	if (value.some(item => typeof item !== 'string' && typeof item !== 'function')) {
		throw new TypeError('Expected array to only contain strings.');
	***REMOVED***
***REMOVED***

//------------------------------------------------------------------------------
// Exports
//------------------------------------------------------------------------------

/**
 * The base schema that every ConfigArray uses.
 * @type Object
 */
const baseSchema = Object.freeze({
	name: {
		required: false,
		merge() {
			return undefined;
		***REMOVED***,
		validate(value) {
			if (typeof value !== 'string') {
				throw new TypeError('Property must be a string.');
			***REMOVED***
		***REMOVED***
	***REMOVED***,
	files: {
		required: false,
		merge() {
			return undefined;
		***REMOVED***,
		validate(value) {

			// first check if it's an array
			assertIsArray(value);

			// then check each member
			value.forEach(item => {
				if (Array.isArray(item)) {
					assertIsArrayOfStringsAndFunctions(item);
				***REMOVED*** else if (typeof item !== 'string' && typeof item !== 'function') {
					throw new TypeError('Items must be a string, a function, or an array of strings and functions.');
				***REMOVED***
			***REMOVED***);

		***REMOVED***
	***REMOVED***,
	ignores: {
		required: false,
		merge() {
			return undefined;
		***REMOVED***,
		validate: assertIsArrayOfStringsAndFunctions
	***REMOVED***
***REMOVED***);

/**
 * @fileoverview ConfigArray
 * @author Nicholas C. Zakas
 */

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

const debug = createDebug('@hwc/config-array');

const MINIMATCH_OPTIONS = {
	matchBase: true
***REMOVED***;

/**
 * Shorthand for checking if a value is a string.
 * @param {any***REMOVED*** value The value to check.
 * @returns {boolean***REMOVED*** True if a string, false if not. 
 */
function isString(value) {
	return typeof value === 'string';
***REMOVED***

/**
 * Normalizes a `ConfigArray` by flattening it and executing any functions
 * that are found inside.
 * @param {Array***REMOVED*** items The items in a `ConfigArray`.
 * @param {Object***REMOVED*** context The context object to pass into any function
 *      found.
 * @returns {Array***REMOVED*** A flattened array containing only config objects.
 * @throws {TypeError***REMOVED*** When a config function returns a function.
 */
async function normalize(items, context) {

	// TODO: Allow async config functions

	function *flatTraverse(array) {
		for (let item of array) {
			if (typeof item === 'function') {
				item = item(context);
			***REMOVED***

			if (Array.isArray(item)) {
				yield * flatTraverse(item);
			***REMOVED*** else if (typeof item === 'function') {
				throw new TypeError('A config function can only return an object or array.');
			***REMOVED*** else {
				yield item;
			***REMOVED***
		***REMOVED***
	***REMOVED***

	return [...flatTraverse(items)];
***REMOVED***

/**
 * Determines if a given file path is matched by a config. If the config
 * has no `files` field, then it matches; otherwise, if a `files` field
 * is present then we match the globs in `files` and exclude any globs in
 * `ignores`.
 * @param {string***REMOVED*** filePath The absolute file path to check.
 * @param {Object***REMOVED*** config The config object to check.
 * @returns {boolean***REMOVED*** True if the file path is matched by the config,
 *      false if not.
 */
function pathMatches(filePath, basePath, config) {

	// a config without a `files` field always matches
	if (!config.files) {
		return true;
	***REMOVED***

	// if files isn't an array, throw an error
	if (!Array.isArray(config.files) || config.files.length === 0) {
		throw new TypeError('The files key must be a non-empty array.');
	***REMOVED***

	const relativeFilePath = path.relative(basePath, filePath);

	// match both strings and functions
	const match = pattern => {
		if (isString(pattern)) {
			return minimatch(relativeFilePath, pattern, MINIMATCH_OPTIONS);
		***REMOVED***

		if (typeof pattern === 'function') {
			return pattern(filePath);
		***REMOVED***
	***REMOVED***;

	// check for all matches to config.files
	let matches = config.files.some(pattern => {
		if (Array.isArray(pattern)) {
			return pattern.every(match);
		***REMOVED***

		return match(pattern);
	***REMOVED***);

	/*
	 * If the file path matches the config.files patterns, then check to see
	 * if there are any files to ignore.
	 */
	if (matches && config.ignores) {
		matches = !config.ignores.some(pattern => {
			return minimatch(filePath, pattern, MINIMATCH_OPTIONS);
		***REMOVED***);
	***REMOVED***

	return matches;
***REMOVED***

/**
 * Ensures that a ConfigArray has been normalized.
 * @param {ConfigArray***REMOVED*** configArray The ConfigArray to check. 
 * @returns {void***REMOVED***
 * @throws {Error***REMOVED*** When the `ConfigArray` is not normalized.
 */
function assertNormalized(configArray) {
	// TODO: Throw more verbose error
	if (!configArray.isNormalized()) {
		throw new Error('ConfigArray must be normalized to perform this operation.');
	***REMOVED***
***REMOVED***

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

const ConfigArraySymbol = {
	isNormalized: Symbol('isNormalized'),
	configCache: Symbol('configCache'),
	schema: Symbol('schema'),
	finalizeConfig: Symbol('finalizeConfig'),
	preprocessConfig: Symbol('preprocessConfig')
***REMOVED***;

/**
 * Represents an array of config objects and provides method for working with
 * those config objects.
 */
class ConfigArray extends Array {

	/**
	 * Creates a new instance of ConfigArray.
	 * @param {Iterable|Function|Object***REMOVED*** configs An iterable yielding config
	 *      objects, or a config function, or a config object.
	 * @param {string***REMOVED*** [options.basePath=""] The path of the config file
	 * @param {boolean***REMOVED*** [options.normalized=false] Flag indicating if the
	 *      configs have already been normalized.
	 * @param {Object***REMOVED*** [options.schema] The additional schema 
	 *      definitions to use for the ConfigArray schema.
	 */
	constructor(configs, { basePath = '', normalized = false, schema: customSchema ***REMOVED*** = {***REMOVED***) {
		super();

		/**
		 * Tracks if the array has been normalized.
		 * @property isNormalized
		 * @type boolean
		 * @private
		 */
		this[ConfigArraySymbol.isNormalized] = normalized;

		/**
		 * The schema used for validating and merging configs.
		 * @property schema
		 * @type ObjectSchema
		 * @private
		 */
		this[ConfigArraySymbol.schema] = new objectSchema.ObjectSchema({
			...customSchema,
			...baseSchema
		***REMOVED***);

		/**
		 * The path of the config file that this array was loaded from.
		 * This is used to calculate filename matches.
		 * @property basePath
		 * @type string
		 */
		this.basePath = basePath;

		/**
		 * A cache to store calculated configs for faster repeat lookup.
		 * @property configCache
		 * @type Map
		 * @private
		 */
		this[ConfigArraySymbol.configCache] = new Map();

		// load the configs into this array
		if (Array.isArray(configs)) {
			this.push(...configs);
		***REMOVED*** else {
			this.push(configs);
		***REMOVED***

	***REMOVED***

	/**
	 * Prevent normal array methods from creating a new `ConfigArray` instance.
	 * This is to ensure that methods such as `slice()` won't try to create a 
	 * new instance of `ConfigArray` behind the scenes as doing so may throw
	 * an error due to the different constructor signature.
	 * @returns {Function***REMOVED*** The `Array` constructor.
	 */
	static get [Symbol.species]() {
		return Array;
	***REMOVED***

	/**
	 * Returns the `files` globs from every config object in the array.
	 * Negated patterns (those beginning with `!`) are not returned.
	 * This can be used to determine which files will be matched by a
	 * config array or to use as a glob pattern when no patterns are provided
	 * for a command line interface.
	 * @returns {string[]***REMOVED*** An array of string patterns.
	 */
	get files() {

		assertNormalized(this);

		const result = [];

		for (const config of this) {
			if (config.files) {
				config.files.forEach(filePattern => {
					if (Array.isArray(filePattern)) {
						result.push(...filePattern.filter(pattern => {
							return isString(pattern) && !pattern.startsWith('!');
						***REMOVED***));
					***REMOVED*** else if (isString(filePattern) && !filePattern.startsWith('!')) {
						result.push(filePattern);
					***REMOVED***
				***REMOVED***);
			***REMOVED***
		***REMOVED***

		return result;
	***REMOVED***

	/**
	 * Returns the file globs that should always be ignored regardless of
	 * the matching `files` fields in any configs. This is necessary to mimic
	 * the behavior of things like .gitignore and .eslintignore, allowing a
	 * globbing operation to be faster.
	 * @returns {string[]***REMOVED*** An array of string patterns to be ignored.
	 */
	get ignores() {

		assertNormalized(this);

		const result = [];

		for (const config of this) {
			if (config.ignores && !config.files) {
				result.push(...config.ignores.filter(isString));
			***REMOVED***
		***REMOVED***

		return result;
	***REMOVED***

	/**
	 * Indicates if the config array has been normalized.
	 * @returns {boolean***REMOVED*** True if the config array is normalized, false if not.
	 */
	isNormalized() {
		return this[ConfigArraySymbol.isNormalized];
	***REMOVED***

	/**
	 * Normalizes a config array by flattening embedded arrays and executing
	 * config functions.
	 * @param {ConfigContext***REMOVED*** context The context object for config functions.
	 * @returns {ConfigArray***REMOVED*** A new ConfigArray instance that is normalized.
	 */
	async normalize(context = {***REMOVED***) {

		if (!this.isNormalized()) {
			const normalizedConfigs = await normalize(this, context);
			this.length = 0;
			this.push(...normalizedConfigs.map(this[ConfigArraySymbol.preprocessConfig]));
			this[ConfigArraySymbol.isNormalized] = true;

			// prevent further changes
			Object.freeze(this);
		***REMOVED***

		return this;
	***REMOVED***

	/**
	 * Finalizes the state of a config before being cached and returned by
	 * `getConfig()`. Does nothing by default but is provided to be
	 * overridden by subclasses as necessary.
	 * @param {Object***REMOVED*** config The config to finalize.
	 * @returns {Object***REMOVED*** The finalized config.
	 */
	[ConfigArraySymbol.finalizeConfig](config) {
		return config;
	***REMOVED***

	/**
	 * Preprocesses a config during the normalization process. This is the
	 * method to override if you want to convert an array item before it is
	 * validated for the first time. For example, if you want to replace a
	 * string with an object, this is the method to override.
	 * @param {Object***REMOVED*** config The config to preprocess.
	 * @returns {Object***REMOVED*** The config to use in place of the argument.
	 */
	[ConfigArraySymbol.preprocessConfig](config) {
		return config;
	***REMOVED***

	/**
	 * Returns the config object for a given file path.
	 * @param {string***REMOVED*** filePath The complete path of a file to get a config for.
	 * @returns {Object***REMOVED*** The config object for this file.
	 */
	getConfig(filePath) {

		assertNormalized(this);

		// first check the cache to avoid duplicate work
		let finalConfig = this[ConfigArraySymbol.configCache].get(filePath);

		if (finalConfig) {
			return finalConfig;
		***REMOVED***

		// No config found in cache, so calculate a new one

		const matchingConfigs = [];

		for (const config of this) {
			if (pathMatches(filePath, this.basePath, config)) {
				debug(`Matching config found for ${filePath***REMOVED***`);
				matchingConfigs.push(config);
			***REMOVED*** else {
				debug(`No matching config found for ${filePath***REMOVED***`);
			***REMOVED***
		***REMOVED***

		finalConfig = matchingConfigs.reduce((result, config) => {
			return this[ConfigArraySymbol.schema].merge(result, config);
		***REMOVED***, {***REMOVED***, this);

		finalConfig = this[ConfigArraySymbol.finalizeConfig](finalConfig);

		this[ConfigArraySymbol.configCache].set(filePath, finalConfig);

		return finalConfig;
	***REMOVED***

***REMOVED***

exports.ConfigArray = ConfigArray;
exports.ConfigArraySymbol = ConfigArraySymbol;
