// css-tree/lib/definition-syntax/tokenizer.js
var SyntaxError = require('./SyntaxError');

var TAB = 9;
var N = 10;
var F = 12;
var R = 13;
var SPACE = 32;

var Tokenizer = function(str) {
    this.str = str;
    this.pos = 0;
***REMOVED***;

Tokenizer.prototype = {
    charCodeAt: function(pos) {
        return pos < this.str.length ? this.str.charCodeAt(pos) : 0;
    ***REMOVED***,
    charCode: function() {
        return this.charCodeAt(this.pos);
    ***REMOVED***,
    nextCharCode: function() {
        return this.charCodeAt(this.pos + 1);
    ***REMOVED***,
    nextNonWsCode: function(pos) {
        return this.charCodeAt(this.findWsEnd(pos));
    ***REMOVED***,
    findWsEnd: function(pos) {
        for (; pos < this.str.length; pos++) {
            var code = this.str.charCodeAt(pos);
            if (code !== R && code !== N && code !== F && code !== SPACE && code !== TAB) {
                break;
            ***REMOVED***
        ***REMOVED***

        return pos;
    ***REMOVED***,
    substringToPos: function(end) {
        return this.str.substring(this.pos, this.pos = end);
    ***REMOVED***,
    eat: function(code) {
        if (this.charCode() !== code) {
            this.error('Expect `' + String.fromCharCode(code) + '`');
        ***REMOVED***

        this.pos++;
    ***REMOVED***,
    peek: function() {
        return this.pos < this.str.length ? this.str.charAt(this.pos++) : '';
    ***REMOVED***,
    error: function(message) {
        throw new SyntaxError(message, this.str, this.pos);
    ***REMOVED***
***REMOVED***;

module.exports = Tokenizer;
