// @babel/runtime/helpers/esm/wrapAsyncGenerator.js
import OverloadYield from "./OverloadYield.js";
function _wrapAsyncGenerator(e) {
  return function () {
    return new AsyncGenerator(e.apply(this, arguments));
  ***REMOVED***;
***REMOVED***
function AsyncGenerator(e) {
  var r, t;
  function resume(r, t) {
    try {
      var n = e[r](t),
        o = n.value,
        u = o instanceof OverloadYield;
      Promise.resolve(u ? o.v : o).then(function (t) {
        if (u) {
          var i = "return" === r ? "return" : "next";
          if (!o.k || t.done) return resume(i, t);
          t = e[i](t).value;
        ***REMOVED***
        settle(n.done ? "return" : "normal", t);
      ***REMOVED***, function (e) {
        resume("throw", e);
      ***REMOVED***);
    ***REMOVED*** catch (e) {
      settle("throw", e);
    ***REMOVED***
  ***REMOVED***
  function settle(e, n) {
    switch (e) {
      case "return":
        r.resolve({
          value: n,
          done: !0
        ***REMOVED***);
        break;
      case "throw":
        r.reject(n);
        break;
      default:
        r.resolve({
          value: n,
          done: !1
        ***REMOVED***);
    ***REMOVED***
    (r = r.next) ? resume(r.key, r.arg) : t = null;
  ***REMOVED***
  this._invoke = function (e, n) {
    return new Promise(function (o, u) {
      var i = {
        key: e,
        arg: n,
        resolve: o,
        reject: u,
        next: null
      ***REMOVED***;
      t ? t = t.next = i : (r = t = i, resume(e, n));
    ***REMOVED***);
  ***REMOVED***, "function" != typeof e["return"] && (this["return"] = void 0);
***REMOVED***
AsyncGenerator.prototype["function" == typeof Symbol && Symbol.asyncIterator || "@@asyncIterator"] = function () {
  return this;
***REMOVED***, AsyncGenerator.prototype.next = function (e) {
  return this._invoke("next", e);
***REMOVED***, AsyncGenerator.prototype["throw"] = function (e) {
  return this._invoke("throw", e);
***REMOVED***, AsyncGenerator.prototype["return"] = function (e) {
  return this._invoke("return", e);
***REMOVED***;
export { _wrapAsyncGenerator as default ***REMOVED***;
