// ajv/lib/compile/codegen/scope.ts
import {_, nil, Code, Name***REMOVED*** from "./code"

interface NameGroup {
  prefix: string
  index: number
***REMOVED***

export interface NameValue {
  ref: ValueReference // this is the reference to any value that can be referred to from generated code via `globals` var in the closure
  key?: unknown // any key to identify a global to avoid duplicates, if not passed ref is used
  code?: Code // this is the code creating the value needed for standalone code wit_out closure - can be a primitive value, function or import (`require`)
***REMOVED***

export type ValueReference = unknown // possibly make CodeGen parameterized type on this type

class ValueError extends Error {
  readonly value?: NameValue
  constructor(name: ValueScopeName) {
    super(`CodeGen: "code" for ${name***REMOVED*** not defined`)
    this.value = name.value
  ***REMOVED***
***REMOVED***

interface ScopeOptions {
  prefixes?: Set<string>
  parent?: Scope
***REMOVED***

interface ValueScopeOptions extends ScopeOptions {
  scope: ScopeStore
  es5?: boolean
  lines?: boolean
***REMOVED***

export type ScopeStore = Record<string, ValueReference[] | undefined>

type ScopeValues = {
  [Prefix in string]?: Map<unknown, ValueScopeName>
***REMOVED***

export type ScopeValueSets = {
  [Prefix in string]?: Set<ValueScopeName>
***REMOVED***

export enum UsedValueState {
  Started,
  Completed,
***REMOVED***

export type UsedScopeValues = {
  [Prefix in string]?: Map<ValueScopeName, UsedValueState | undefined>
***REMOVED***

export const varKinds = {
  const: new Name("const"),
  let: new Name("let"),
  var: new Name("var"),
***REMOVED***

export class Scope {
  protected readonly _names: {[Prefix in string]?: NameGroup***REMOVED*** = {***REMOVED***
  protected readonly _prefixes?: Set<string>
  protected readonly _parent?: Scope

  constructor({prefixes, parent***REMOVED***: ScopeOptions = {***REMOVED***) {
    this._prefixes = prefixes
    this._parent = parent
  ***REMOVED***

  toName(nameOrPrefix: Name | string): Name {
    return nameOrPrefix instanceof Name ? nameOrPrefix : this.name(nameOrPrefix)
  ***REMOVED***

  name(prefix: string): Name {
    return new Name(this._newName(prefix))
  ***REMOVED***

  protected _newName(prefix: string): string {
    const ng = this._names[prefix] || this._nameGroup(prefix)
    return `${prefix***REMOVED***${ng.index++***REMOVED***`
  ***REMOVED***

  private _nameGroup(prefix: string): NameGroup {
    if (this._parent?._prefixes?.has(prefix) || (this._prefixes && !this._prefixes.has(prefix))) {
      throw new Error(`CodeGen: prefix "${prefix***REMOVED***" is not allowed in this scope`)
    ***REMOVED***
    return (this._names[prefix] = {prefix, index: 0***REMOVED***)
  ***REMOVED***
***REMOVED***

interface ScopePath {
  property: string
  itemIndex: number
***REMOVED***

export class ValueScopeName extends Name {
  readonly prefix: string
  value?: NameValue
  scopePath?: Code

  constructor(prefix: string, nameStr: string) {
    super(nameStr)
    this.prefix = prefix
  ***REMOVED***

  setValue(value: NameValue, {property, itemIndex***REMOVED***: ScopePath): void {
    this.value = value
    this.scopePath = _`.${new Name(property)***REMOVED***[${itemIndex***REMOVED***]`
  ***REMOVED***
***REMOVED***

interface VSOptions extends ValueScopeOptions {
  _n: Code
***REMOVED***

const line = _`\n`

export class ValueScope extends Scope {
  protected readonly _values: ScopeValues = {***REMOVED***
  protected readonly _scope: ScopeStore
  readonly opts: VSOptions

  constructor(opts: ValueScopeOptions) {
    super(opts)
    this._scope = opts.scope
    this.opts = {...opts, _n: opts.lines ? line : nil***REMOVED***
  ***REMOVED***

  get(): ScopeStore {
    return this._scope
  ***REMOVED***

  name(prefix: string): ValueScopeName {
    return new ValueScopeName(prefix, this._newName(prefix))
  ***REMOVED***

  value(nameOrPrefix: ValueScopeName | string, value: NameValue): ValueScopeName {
    if (value.ref === undefined) throw new Error("CodeGen: ref must be passed in value")
    const name = this.toName(nameOrPrefix) as ValueScopeName
    const {prefix***REMOVED*** = name
    const valueKey = value.key ?? value.ref
    let vs = this._values[prefix]
    if (vs) {
      const _name = vs.get(valueKey)
      if (_name) return _name
    ***REMOVED*** else {
      vs = this._values[prefix] = new Map()
    ***REMOVED***
    vs.set(valueKey, name)

    const s = this._scope[prefix] || (this._scope[prefix] = [])
    const itemIndex = s.length
    s[itemIndex] = value.ref
    name.setValue(value, {property: prefix, itemIndex***REMOVED***)
    return name
  ***REMOVED***

  getValue(prefix: string, keyOrRef: unknown): ValueScopeName | undefined {
    const vs = this._values[prefix]
    if (!vs) return
    return vs.get(keyOrRef)
  ***REMOVED***

  scopeRefs(scopeName: Name, values: ScopeValues | ScopeValueSets = this._values): Code {
    return this._reduceValues(values, (name: ValueScopeName) => {
      if (name.scopePath === undefined) throw new Error(`CodeGen: name "${name***REMOVED***" has no value`)
      return _`${scopeName***REMOVED***${name.scopePath***REMOVED***`
    ***REMOVED***)
  ***REMOVED***

  scopeCode(
    values: ScopeValues | ScopeValueSets = this._values,
    usedValues?: UsedScopeValues,
    getCode?: (n: ValueScopeName) => Code | undefined
  ): Code {
    return this._reduceValues(
      values,
      (name: ValueScopeName) => {
        if (name.value === undefined) throw new Error(`CodeGen: name "${name***REMOVED***" has no value`)
        return name.value.code
      ***REMOVED***,
      usedValues,
      getCode
    )
  ***REMOVED***

  private _reduceValues(
    values: ScopeValues | ScopeValueSets,
    valueCode: (n: ValueScopeName) => Code | undefined,
    usedValues: UsedScopeValues = {***REMOVED***,
    getCode?: (n: ValueScopeName) => Code | undefined
  ): Code {
    let code: Code = nil
    for (const prefix in values) {
      const vs = values[prefix]
      if (!vs) continue
      const nameSet = (usedValues[prefix] = usedValues[prefix] || new Map())
      vs.forEach((name: ValueScopeName) => {
        if (nameSet.has(name)) return
        nameSet.set(name, UsedValueState.Started)
        let c = valueCode(name)
        if (c) {
          const def = this.opts.es5 ? varKinds.var : varKinds.const
          code = _`${code***REMOVED***${def***REMOVED*** ${name***REMOVED*** = ${c***REMOVED***;${this.opts._n***REMOVED***`
        ***REMOVED*** else if ((c = getCode?.(name))) {
          code = _`${code***REMOVED***${c***REMOVED***${this.opts._n***REMOVED***`
        ***REMOVED*** else {
          throw new ValueError(name)
        ***REMOVED***
        nameSet.set(name, UsedValueState.Completed)
      ***REMOVED***)
    ***REMOVED***
    return code
  ***REMOVED***
***REMOVED***
