// istanbul-reports/lib/json-summary/index.js
/*
 Copyright 2012-2015, Yahoo Inc.
 Copyrights licensed under the New BSD License. See the accompanying LICENSE file for terms.
 */
'use strict';
const { ReportBase ***REMOVED*** = require('istanbul-lib-report');

class JsonSummaryReport extends ReportBase {
    constructor(opts) {
        super();

        this.file = opts.file || 'coverage-summary.json';
        this.contentWriter = null;
        this.first = true;
    ***REMOVED***

    onStart(root, context) {
        this.contentWriter = context.writer.writeFile(this.file);
        this.contentWriter.write('{');
    ***REMOVED***

    writeSummary(filePath, sc) {
        const cw = this.contentWriter;
        if (this.first) {
            this.first = false;
        ***REMOVED*** else {
            cw.write(',');
        ***REMOVED***
        cw.write(JSON.stringify(filePath));
        cw.write(': ');
        cw.write(JSON.stringify(sc));
        cw.println('');
    ***REMOVED***

    onSummary(node) {
        if (!node.isRoot()) {
            return;
        ***REMOVED***
        this.writeSummary('total', node.getCoverageSummary());
    ***REMOVED***

    onDetail(node) {
        this.writeSummary(
            node.getFileCoverage().path,
            node.getCoverageSummary()
        );
    ***REMOVED***

    onEnd() {
        const cw = this.contentWriter;
        cw.println('***REMOVED***');
        cw.close();
    ***REMOVED***
***REMOVED***

module.exports = JsonSummaryReport;
