// autoprefixer/lib/hacks/display-flex.js
let OldValue = require('../old-value')
let Value = require('../value')
let flexSpec = require('./flex-spec')

class DisplayFlex extends Value {
  constructor(name, prefixes) {
    super(name, prefixes)
    if (name === 'display-flex') {
      this.name = 'flex'
    ***REMOVED***
  ***REMOVED***

  /**
   * Faster check for flex value
   */
  check(decl) {
    return decl.prop === 'display' && decl.value === this.name
  ***REMOVED***

  /**
   * Change value for old specs
   */
  old(prefix) {
    let prefixed = this.prefixed(prefix)
    if (!prefixed) return undefined
    return new OldValue(this.name, prefixed)
  ***REMOVED***

  /**
   * Return value by spec
   */
  prefixed(prefix) {
    let spec, value
    ;[spec, prefix] = flexSpec(prefix)

    if (spec === 2009) {
      if (this.name === 'flex') {
        value = 'box'
      ***REMOVED*** else {
        value = 'inline-box'
      ***REMOVED***
    ***REMOVED*** else if (spec === 2012) {
      if (this.name === 'flex') {
        value = 'flexbox'
      ***REMOVED*** else {
        value = 'inline-flexbox'
      ***REMOVED***
    ***REMOVED*** else if (spec === 'final') {
      value = this.name
    ***REMOVED***

    return prefix + value
  ***REMOVED***

  /**
   * Add prefix to value depend on flebox spec version
   */
  replace(string, prefix) {
    return this.prefixed(prefix)
  ***REMOVED***
***REMOVED***

DisplayFlex.names = ['display-flex', 'inline-flex']

module.exports = DisplayFlex
