// autoprefixer/lib/hacks/gradient.js
let range = require('normalize-range')
let parser = require('postcss-value-parser')

let OldValue = require('../old-value')
let utils = require('../utils')
let Value = require('../value')

let IS_DIRECTION = /top|left|right|bottom/gi

class Gradient extends Value {
  /**
   * Do not add non-webkit prefixes for list-style and object
   */
  add(decl, prefix) {
    let p = decl.prop
    if (p.includes('mask')) {
      if (prefix === '-webkit-' || prefix === '-webkit- old') {
        return super.add(decl, prefix)
      ***REMOVED***
    ***REMOVED*** else if (
      p === 'list-style' ||
      p === 'list-style-image' ||
      p === 'content'
    ) {
      if (prefix === '-webkit-' || prefix === '-webkit- old') {
        return super.add(decl, prefix)
      ***REMOVED***
    ***REMOVED*** else {
      return super.add(decl, prefix)
    ***REMOVED***
    return undefined
  ***REMOVED***

  /**
   * Get div token from exists parameters
   */
  cloneDiv(params) {
    for (let i of params) {
      if (i.type === 'div' && i.value === ',') {
        return i
      ***REMOVED***
    ***REMOVED***
    return { after: ' ', type: 'div', value: ',' ***REMOVED***
  ***REMOVED***

  /**
   * Change colors syntax to old webkit
   */
  colorStops(params) {
    let result = []
    for (let i = 0; i < params.length; i++) {
      let pos
      let param = params[i]
      let item
      if (i === 0) {
        continue
      ***REMOVED***

      let color = parser.stringify(param[0])
      if (param[1] && param[1].type === 'word') {
        pos = param[1].value
      ***REMOVED*** else if (param[2] && param[2].type === 'word') {
        pos = param[2].value
      ***REMOVED***

      let stop
      if (i === 1 && (!pos || pos === '0%')) {
        stop = `from(${color***REMOVED***)`
      ***REMOVED*** else if (i === params.length - 1 && (!pos || pos === '100%')) {
        stop = `to(${color***REMOVED***)`
      ***REMOVED*** else if (pos) {
        stop = `color-stop(${pos***REMOVED***, ${color***REMOVED***)`
      ***REMOVED*** else {
        stop = `color-stop(${color***REMOVED***)`
      ***REMOVED***

      let div = param[param.length - 1]
      params[i] = [{ type: 'word', value: stop ***REMOVED***]
      if (div.type === 'div' && div.value === ',') {
        item = params[i].push(div)
      ***REMOVED***
      result.push(item)
    ***REMOVED***
    return result
  ***REMOVED***

  /**
   * Change new direction to old
   */
  convertDirection(params) {
    if (params.length > 0) {
      if (params[0].value === 'to') {
        this.fixDirection(params)
      ***REMOVED*** else if (params[0].value.includes('deg')) {
        this.fixAngle(params)
      ***REMOVED*** else if (this.isRadial(params)) {
        this.fixRadial(params)
      ***REMOVED***
    ***REMOVED***
    return params
  ***REMOVED***

  /**
   * Add 90 degrees
   */
  fixAngle(params) {
    let first = params[0].value
    first = parseFloat(first)
    first = Math.abs(450 - first) % 360
    first = this.roundFloat(first, 3)
    params[0].value = `${first***REMOVED***deg`
  ***REMOVED***

  /**
   * Replace `to top left` to `bottom right`
   */
  fixDirection(params) {
    params.splice(0, 2)

    for (let param of params) {
      if (param.type === 'div') {
        break
      ***REMOVED***
      if (param.type === 'word') {
        param.value = this.revertDirection(param.value)
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  /**
   * Fix radial direction syntax
   */
  fixRadial(params) {
    let first = []
    let second = []
    let a, b, c, i, next

    for (i = 0; i < params.length - 2; i++) {
      a = params[i]
      b = params[i + 1]
      c = params[i + 2]
      if (a.type === 'space' && b.value === 'at' && c.type === 'space') {
        next = i + 3
        break
      ***REMOVED*** else {
        first.push(a)
      ***REMOVED***
    ***REMOVED***

    let div
    for (i = next; i < params.length; i++) {
      if (params[i].type === 'div') {
        div = params[i]
        break
      ***REMOVED*** else {
        second.push(params[i])
      ***REMOVED***
    ***REMOVED***

    params.splice(0, i, ...second, div, ...first)
  ***REMOVED***

  /**
   * Look for at word
   */
  isRadial(params) {
    let state = 'before'
    for (let param of params) {
      if (state === 'before' && param.type === 'space') {
        state = 'at'
      ***REMOVED*** else if (state === 'at' && param.value === 'at') {
        state = 'after'
      ***REMOVED*** else if (state === 'after' && param.type === 'space') {
        return true
      ***REMOVED*** else if (param.type === 'div') {
        break
      ***REMOVED*** else {
        state = 'before'
      ***REMOVED***
    ***REMOVED***
    return false
  ***REMOVED***

  /**
   * Replace old direction to new
   */
  newDirection(params) {
    if (params[0].value === 'to') {
      return params
    ***REMOVED***
    IS_DIRECTION.lastIndex = 0 // reset search index of global regexp
    if (!IS_DIRECTION.test(params[0].value)) {
      return params
    ***REMOVED***

    params.unshift(
      {
        type: 'word',
        value: 'to'
      ***REMOVED***,
      {
        type: 'space',
        value: ' '
      ***REMOVED***
    )

    for (let i = 2; i < params.length; i++) {
      if (params[i].type === 'div') {
        break
      ***REMOVED***
      if (params[i].type === 'word') {
        params[i].value = this.revertDirection(params[i].value)
      ***REMOVED***
    ***REMOVED***

    return params
  ***REMOVED***

  /**
   * Normalize angle
   */
  normalize(nodes, gradientName) {
    if (!nodes[0]) return nodes

    if (/-?\d+(.\d+)?grad/.test(nodes[0].value)) {
      nodes[0].value = this.normalizeUnit(nodes[0].value, 400)
    ***REMOVED*** else if (/-?\d+(.\d+)?rad/.test(nodes[0].value)) {
      nodes[0].value = this.normalizeUnit(nodes[0].value, 2 * Math.PI)
    ***REMOVED*** else if (/-?\d+(.\d+)?turn/.test(nodes[0].value)) {
      nodes[0].value = this.normalizeUnit(nodes[0].value, 1)
    ***REMOVED*** else if (nodes[0].value.includes('deg')) {
      let num = parseFloat(nodes[0].value)
      num = range.wrap(0, 360, num)
      nodes[0].value = `${num***REMOVED***deg`
    ***REMOVED***

    if (
      gradientName === 'linear-gradient' ||
      gradientName === 'repeating-linear-gradient'
    ) {
      let direction = nodes[0].value

      // Unitless zero for `<angle>` values are allowed in CSS gradients and transforms.
      // Spec: https://github.com/w3c/csswg-drafts/commit/602789171429b2231223ab1e5acf8f7f11652eb3
      if (direction === '0deg' || direction === '0') {
        nodes = this.replaceFirst(nodes, 'to', ' ', 'top')
      ***REMOVED*** else if (direction === '90deg') {
        nodes = this.replaceFirst(nodes, 'to', ' ', 'right')
      ***REMOVED*** else if (direction === '180deg') {
        nodes = this.replaceFirst(nodes, 'to', ' ', 'bottom') // default value
      ***REMOVED*** else if (direction === '270deg') {
        nodes = this.replaceFirst(nodes, 'to', ' ', 'left')
      ***REMOVED***
    ***REMOVED***

    return nodes
  ***REMOVED***

  /**
   * Convert angle unit to deg
   */
  normalizeUnit(str, full) {
    let num = parseFloat(str)
    let deg = (num / full) * 360
    return `${deg***REMOVED***deg`
  ***REMOVED***

  /**
   * Remove old WebKit gradient too
   */
  old(prefix) {
    if (prefix === '-webkit-') {
      let type
      if (this.name === 'linear-gradient') {
        type = 'linear'
      ***REMOVED*** else if (this.name === 'repeating-linear-gradient') {
        type = 'repeating-linear'
      ***REMOVED*** else if (this.name === 'repeating-radial-gradient') {
        type = 'repeating-radial'
      ***REMOVED*** else {
        type = 'radial'
      ***REMOVED***
      let string = '-gradient'
      let regexp = utils.regexp(
        `-webkit-(${type***REMOVED***-gradient|gradient\\(\\s*${type***REMOVED***)`,
        false
      )

      return new OldValue(this.name, prefix + this.name, string, regexp)
    ***REMOVED*** else {
      return super.old(prefix)
    ***REMOVED***
  ***REMOVED***

  /**
   * Change direction syntax to old webkit
   */
  oldDirection(params) {
    let div = this.cloneDiv(params[0])

    if (params[0][0].value !== 'to') {
      return params.unshift([
        { type: 'word', value: Gradient.oldDirections.bottom ***REMOVED***,
        div
      ])
    ***REMOVED*** else {
      let words = []
      for (let node of params[0].slice(2)) {
        if (node.type === 'word') {
          words.push(node.value.toLowerCase())
        ***REMOVED***
      ***REMOVED***

      words = words.join(' ')
      let old = Gradient.oldDirections[words] || words

      params[0] = [{ type: 'word', value: old ***REMOVED***, div]
      return params[0]
    ***REMOVED***
  ***REMOVED***

  /**
   * Convert to old webkit syntax
   */
  oldWebkit(node) {
    let { nodes ***REMOVED*** = node
    let string = parser.stringify(node.nodes)

    if (this.name !== 'linear-gradient') {
      return false
    ***REMOVED***
    if (nodes[0] && nodes[0].value.includes('deg')) {
      return false
    ***REMOVED***
    if (
      string.includes('px') ||
      string.includes('-corner') ||
      string.includes('-side')
    ) {
      return false
    ***REMOVED***

    let params = [[]]
    for (let i of nodes) {
      params[params.length - 1].push(i)
      if (i.type === 'div' && i.value === ',') {
        params.push([])
      ***REMOVED***
    ***REMOVED***

    this.oldDirection(params)
    this.colorStops(params)

    node.nodes = []
    for (let param of params) {
      node.nodes = node.nodes.concat(param)
    ***REMOVED***

    node.nodes.unshift(
      { type: 'word', value: 'linear' ***REMOVED***,
      this.cloneDiv(node.nodes)
    )
    node.value = '-webkit-gradient'

    return true
  ***REMOVED***

  /**
   * Change degrees for webkit prefix
   */
  replace(string, prefix) {
    let ast = parser(string)
    for (let node of ast.nodes) {
      let gradientName = this.name // gradient name
      if (node.type === 'function' && node.value === gradientName) {
        node.nodes = this.newDirection(node.nodes)
        node.nodes = this.normalize(node.nodes, gradientName)
        if (prefix === '-webkit- old') {
          let changes = this.oldWebkit(node)
          if (!changes) {
            return false
          ***REMOVED***
        ***REMOVED*** else {
          node.nodes = this.convertDirection(node.nodes)
          node.value = prefix + node.value
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
    return ast.toString()
  ***REMOVED***

  /**
   * Replace first token
   */
  replaceFirst(params, ...words) {
    let prefix = words.map(i => {
      if (i === ' ') {
        return { type: 'space', value: i ***REMOVED***
      ***REMOVED***
      return { type: 'word', value: i ***REMOVED***
    ***REMOVED***)
    return prefix.concat(params.slice(1))
  ***REMOVED***

  revertDirection(word) {
    return Gradient.directions[word.toLowerCase()] || word
  ***REMOVED***

  /**
   * Round float and save digits under dot
   */
  roundFloat(float, digits) {
    return parseFloat(float.toFixed(digits))
  ***REMOVED***
***REMOVED***

Gradient.names = [
  'linear-gradient',
  'repeating-linear-gradient',
  'radial-gradient',
  'repeating-radial-gradient'
]

Gradient.directions = {
  bottom: 'top',
  left: 'right',
  right: 'left',
  top: 'bottom' // default value
***REMOVED***

// Direction to replace
Gradient.oldDirections = {
  'bottom': 'left top, left bottom',
  'bottom left': 'right top, left bottom',
  'bottom right': 'left top, right bottom',
  'left': 'right top, left top',

  'left bottom': 'right top, left bottom',
  'left top': 'right bottom, left top',
  'right': 'left top, right top',
  'right bottom': 'left top, right bottom',
  'right top': 'left bottom, right top',
  'top': 'left bottom, left top',
  'top left': 'right bottom, left top',
  'top right': 'left bottom, right top'
***REMOVED***

module.exports = Gradient
