// autoprefixer/lib/hacks/grid-utils.js
let parser = require('postcss-value-parser')
let list = require('postcss').list

let uniq = require('../utils').uniq
let escapeRegexp = require('../utils').escapeRegexp
let splitSelector = require('../utils').splitSelector

function convert(value) {
  if (
    value &&
    value.length === 2 &&
    value[0] === 'span' &&
    parseInt(value[1], 10) > 0
  ) {
    return [false, parseInt(value[1], 10)]
  ***REMOVED***

  if (value && value.length === 1 && parseInt(value[0], 10) > 0) {
    return [parseInt(value[0], 10), false]
  ***REMOVED***

  return [false, false]
***REMOVED***

exports.translate = translate

function translate(values, startIndex, endIndex) {
  let startValue = values[startIndex]
  let endValue = values[endIndex]

  if (!startValue) {
    return [false, false]
  ***REMOVED***

  let [start, spanStart] = convert(startValue)
  let [end, spanEnd] = convert(endValue)

  if (start && !endValue) {
    return [start, false]
  ***REMOVED***

  if (spanStart && end) {
    return [end - spanStart, spanStart]
  ***REMOVED***

  if (start && spanEnd) {
    return [start, spanEnd]
  ***REMOVED***

  if (start && end) {
    return [start, end - start]
  ***REMOVED***

  return [false, false]
***REMOVED***

exports.parse = parse

function parse(decl) {
  let node = parser(decl.value)

  let values = []
  let current = 0
  values[current] = []

  for (let i of node.nodes) {
    if (i.type === 'div') {
      current += 1
      values[current] = []
    ***REMOVED*** else if (i.type === 'word') {
      values[current].push(i.value)
    ***REMOVED***
  ***REMOVED***

  return values
***REMOVED***

exports.insertDecl = insertDecl

function insertDecl(decl, prop, value) {
  if (value && !decl.parent.some(i => i.prop === `-ms-${prop***REMOVED***`)) {
    decl.cloneBefore({
      prop: `-ms-${prop***REMOVED***`,
      value: value.toString()
    ***REMOVED***)
  ***REMOVED***
***REMOVED***

// Track transforms

exports.prefixTrackProp = prefixTrackProp

function prefixTrackProp({ prefix, prop ***REMOVED***) {
  return prefix + prop.replace('template-', '')
***REMOVED***

function transformRepeat({ nodes ***REMOVED***, { gap ***REMOVED***) {
  let { count, size ***REMOVED*** = nodes.reduce(
    (result, node) => {
      if (node.type === 'div' && node.value === ',') {
        result.key = 'size'
      ***REMOVED*** else {
        result[result.key].push(parser.stringify(node))
      ***REMOVED***
      return result
    ***REMOVED***,
    {
      count: [],
      key: 'count',
      size: []
    ***REMOVED***
  )

  // insert gap values
  if (gap) {
    size = size.filter(i => i.trim())
    let val = []
    for (let i = 1; i <= count; i++) {
      size.forEach((item, index) => {
        if (index > 0 || i > 1) {
          val.push(gap)
        ***REMOVED***
        val.push(item)
      ***REMOVED***)
    ***REMOVED***

    return val.join(' ')
  ***REMOVED***

  return `(${size.join('')***REMOVED***)[${count.join('')***REMOVED***]`
***REMOVED***

exports.prefixTrackValue = prefixTrackValue

function prefixTrackValue({ gap, value ***REMOVED***) {
  let result = parser(value).nodes.reduce((nodes, node) => {
    if (node.type === 'function' && node.value === 'repeat') {
      return nodes.concat({
        type: 'word',
        value: transformRepeat(node, { gap ***REMOVED***)
      ***REMOVED***)
    ***REMOVED***
    if (gap && node.type === 'space') {
      return nodes.concat(
        {
          type: 'space',
          value: ' '
        ***REMOVED***,
        {
          type: 'word',
          value: gap
        ***REMOVED***,
        node
      )
    ***REMOVED***
    return nodes.concat(node)
  ***REMOVED***, [])

  return parser.stringify(result)
***REMOVED***

// Parse grid-template-areas

let DOTS = /^\.+$/

function track(start, end) {
  return { end, span: end - start, start ***REMOVED***
***REMOVED***

function getColumns(line) {
  return line.trim().split(/\s+/g)
***REMOVED***

exports.parseGridAreas = parseGridAreas

function parseGridAreas({ gap, rows ***REMOVED***) {
  return rows.reduce((areas, line, rowIndex) => {
    if (gap.row) rowIndex *= 2

    if (line.trim() === '') return areas

    getColumns(line).forEach((area, columnIndex) => {
      if (DOTS.test(area)) return

      if (gap.column) columnIndex *= 2

      if (typeof areas[area] === 'undefined') {
        areas[area] = {
          column: track(columnIndex + 1, columnIndex + 2),
          row: track(rowIndex + 1, rowIndex + 2)
        ***REMOVED***
      ***REMOVED*** else {
        let { column, row ***REMOVED*** = areas[area]

        column.start = Math.min(column.start, columnIndex + 1)
        column.end = Math.max(column.end, columnIndex + 2)
        column.span = column.end - column.start

        row.start = Math.min(row.start, rowIndex + 1)
        row.end = Math.max(row.end, rowIndex + 2)
        row.span = row.end - row.start
      ***REMOVED***
    ***REMOVED***)

    return areas
  ***REMOVED***, {***REMOVED***)
***REMOVED***

// Parse grid-template

function testTrack(node) {
  return node.type === 'word' && /^\[.+]$/.test(node.value)
***REMOVED***

function verifyRowSize(result) {
  if (result.areas.length > result.rows.length) {
    result.rows.push('auto')
  ***REMOVED***
  return result
***REMOVED***

exports.parseTemplate = parseTemplate

function parseTemplate({ decl, gap ***REMOVED***) {
  let gridTemplate = parser(decl.value).nodes.reduce(
    (result, node) => {
      let { type, value ***REMOVED*** = node

      if (testTrack(node) || type === 'space') return result

      // area
      if (type === 'string') {
        result = verifyRowSize(result)
        result.areas.push(value)
      ***REMOVED***

      // values and function
      if (type === 'word' || type === 'function') {
        result[result.key].push(parser.stringify(node))
      ***REMOVED***

      // divider(/)
      if (type === 'div' && value === '/') {
        result.key = 'columns'
        result = verifyRowSize(result)
      ***REMOVED***

      return result
    ***REMOVED***,
    {
      areas: [],
      columns: [],
      key: 'rows',
      rows: []
    ***REMOVED***
  )

  return {
    areas: parseGridAreas({
      gap,
      rows: gridTemplate.areas
    ***REMOVED***),
    columns: prefixTrackValue({
      gap: gap.column,
      value: gridTemplate.columns.join(' ')
    ***REMOVED***),
    rows: prefixTrackValue({
      gap: gap.row,
      value: gridTemplate.rows.join(' ')
    ***REMOVED***)
  ***REMOVED***
***REMOVED***

// Insert parsed grid areas

/**
 * Get an array of -ms- prefixed props and values
 * @param  {Object***REMOVED*** [area] area object with column and row data
 * @param  {Boolean***REMOVED*** [addRowSpan] should we add grid-column-row value?
 * @param  {Boolean***REMOVED*** [addColumnSpan] should we add grid-column-span value?
 * @return {Array<Object>***REMOVED***
 */
function getMSDecls(area, addRowSpan = false, addColumnSpan = false) {
  let result = [
    {
      prop: '-ms-grid-row',
      value: String(area.row.start)
    ***REMOVED***
  ]
  if (area.row.span > 1 || addRowSpan) {
    result.push({
      prop: '-ms-grid-row-span',
      value: String(area.row.span)
    ***REMOVED***)
  ***REMOVED***
  result.push({
    prop: '-ms-grid-column',
    value: String(area.column.start)
  ***REMOVED***)
  if (area.column.span > 1 || addColumnSpan) {
    result.push({
      prop: '-ms-grid-column-span',
      value: String(area.column.span)
    ***REMOVED***)
  ***REMOVED***
  return result
***REMOVED***

function getParentMedia(parent) {
  if (parent.type === 'atrule' && parent.name === 'media') {
    return parent
  ***REMOVED***
  if (!parent.parent) {
    return false
  ***REMOVED***
  return getParentMedia(parent.parent)
***REMOVED***

/**
 * change selectors for rules with duplicate grid-areas.
 * @param  {Array<Rule>***REMOVED*** rules
 * @param  {Array<String>***REMOVED*** templateSelectors
 * @return {Array<Rule>***REMOVED*** rules with changed selectors
 */
function changeDuplicateAreaSelectors(ruleSelectors, templateSelectors) {
  ruleSelectors = ruleSelectors.map(selector => {
    let selectorBySpace = list.space(selector)
    let selectorByComma = list.comma(selector)

    if (selectorBySpace.length > selectorByComma.length) {
      selector = selectorBySpace.slice(-1).join('')
    ***REMOVED***
    return selector
  ***REMOVED***)

  return ruleSelectors.map(ruleSelector => {
    let newSelector = templateSelectors.map((tplSelector, index) => {
      let space = index === 0 ? '' : ' '
      return `${space***REMOVED***${tplSelector***REMOVED*** > ${ruleSelector***REMOVED***`
    ***REMOVED***)

    return newSelector
  ***REMOVED***)
***REMOVED***

/**
 * check if selector of rules are equal
 * @param  {Rule***REMOVED*** ruleA
 * @param  {Rule***REMOVED*** ruleB
 * @return {Boolean***REMOVED***
 */
function selectorsEqual(ruleA, ruleB) {
  return ruleA.selectors.some(sel => {
    return ruleB.selectors.includes(sel)
  ***REMOVED***)
***REMOVED***

/**
 * Parse data from all grid-template(-areas) declarations
 * @param  {Root***REMOVED*** css css root
 * @return {Object***REMOVED*** parsed data
 */
function parseGridTemplatesData(css) {
  let parsed = []

  // we walk through every grid-template(-areas) declaration and store
  // data with the same area names inside the item
  css.walkDecls(/grid-template(-areas)?$/, d => {
    let rule = d.parent
    let media = getParentMedia(rule)
    let gap = getGridGap(d)
    let inheritedGap = inheritGridGap(d, gap)
    let { areas ***REMOVED*** = parseTemplate({ decl: d, gap: inheritedGap || gap ***REMOVED***)
    let areaNames = Object.keys(areas)

    // skip node if it doesn't have areas
    if (areaNames.length === 0) {
      return true
    ***REMOVED***

    // check parsed array for item that include the same area names
    // return index of that item
    let index = parsed.reduce((acc, { allAreas ***REMOVED***, idx) => {
      let hasAreas = allAreas && areaNames.some(area => allAreas.includes(area))
      return hasAreas ? idx : acc
    ***REMOVED***, null)

    if (index !== null) {
      // index is found, add the grid-template data to that item
      let { allAreas, rules ***REMOVED*** = parsed[index]

      // check if rule has no duplicate area names
      let hasNoDuplicates = rules.some(r => {
        return r.hasDuplicates === false && selectorsEqual(r, rule)
      ***REMOVED***)

      let duplicatesFound = false

      // check need to gather all duplicate area names
      let duplicateAreaNames = rules.reduce((acc, r) => {
        if (!r.params && selectorsEqual(r, rule)) {
          duplicatesFound = true
          return r.duplicateAreaNames
        ***REMOVED***
        if (!duplicatesFound) {
          areaNames.forEach(name => {
            if (r.areas[name]) {
              acc.push(name)
            ***REMOVED***
          ***REMOVED***)
        ***REMOVED***
        return uniq(acc)
      ***REMOVED***, [])

      // update grid-row/column-span values for areas with duplicate
      // area names. @see #1084 and #1146
      rules.forEach(r => {
        areaNames.forEach(name => {
          let area = r.areas[name]
          if (area && area.row.span !== areas[name].row.span) {
            areas[name].row.updateSpan = true
          ***REMOVED***

          if (area && area.column.span !== areas[name].column.span) {
            areas[name].column.updateSpan = true
          ***REMOVED***
        ***REMOVED***)
      ***REMOVED***)

      parsed[index].allAreas = uniq([...allAreas, ...areaNames])
      parsed[index].rules.push({
        areas,
        duplicateAreaNames,
        hasDuplicates: !hasNoDuplicates,
        node: rule,
        params: media.params,
        selectors: rule.selectors
      ***REMOVED***)
    ***REMOVED*** else {
      // index is NOT found, push the new item to the parsed array
      parsed.push({
        allAreas: areaNames,
        areasCount: 0,
        rules: [
          {
            areas,
            duplicateAreaNames: [],
            duplicateRules: [],
            hasDuplicates: false,
            node: rule,
            params: media.params,
            selectors: rule.selectors
          ***REMOVED***
        ]
      ***REMOVED***)
    ***REMOVED***

    return undefined
  ***REMOVED***)

  return parsed
***REMOVED***

/**
 * insert prefixed grid-area declarations
 * @param  {Root***REMOVED***  css css root
 * @param  {Function***REMOVED*** isDisabled check if the rule is disabled
 * @return {void***REMOVED***
 */
exports.insertAreas = insertAreas

function insertAreas(css, isDisabled) {
  // parse grid-template declarations
  let gridTemplatesData = parseGridTemplatesData(css)

  // return undefined if no declarations found
  if (gridTemplatesData.length === 0) {
    return undefined
  ***REMOVED***

  // we need to store the rules that we will insert later
  let rulesToInsert = {***REMOVED***

  css.walkDecls('grid-area', gridArea => {
    let gridAreaRule = gridArea.parent
    let hasPrefixedRow = gridAreaRule.first.prop === '-ms-grid-row'
    let gridAreaMedia = getParentMedia(gridAreaRule)

    if (isDisabled(gridArea)) {
      return undefined
    ***REMOVED***

    let gridAreaRuleIndex = css.index(gridAreaMedia || gridAreaRule)

    let value = gridArea.value
    // found the data that matches grid-area identifier
    let data = gridTemplatesData.filter(d => d.allAreas.includes(value))[0]

    if (!data) {
      return true
    ***REMOVED***

    let lastArea = data.allAreas[data.allAreas.length - 1]
    let selectorBySpace = list.space(gridAreaRule.selector)
    let selectorByComma = list.comma(gridAreaRule.selector)
    let selectorIsComplex =
      selectorBySpace.length > 1 &&
      selectorBySpace.length > selectorByComma.length

    // prevent doubling of prefixes
    if (hasPrefixedRow) {
      return false
    ***REMOVED***

    // create the empty object with the key as the last area name
    // e.g if we have templates with "a b c" values, "c" will be the last area
    if (!rulesToInsert[lastArea]) {
      rulesToInsert[lastArea] = {***REMOVED***
    ***REMOVED***

    let lastRuleIsSet = false

    // walk through every grid-template rule data
    for (let rule of data.rules) {
      let area = rule.areas[value]
      let hasDuplicateName = rule.duplicateAreaNames.includes(value)

      // if we can't find the area name, update lastRule and continue
      if (!area) {
        let lastRule = rulesToInsert[lastArea].lastRule
        let lastRuleIndex
        if (lastRule) {
          lastRuleIndex = css.index(lastRule)
        ***REMOVED*** else {
          /* c8 ignore next 2 */
          lastRuleIndex = -1
        ***REMOVED***

        if (gridAreaRuleIndex > lastRuleIndex) {
          rulesToInsert[lastArea].lastRule = gridAreaMedia || gridAreaRule
        ***REMOVED***
        continue
      ***REMOVED***

      // for grid-templates inside media rule we need to create empty
      // array to push prefixed grid-area rules later
      if (rule.params && !rulesToInsert[lastArea][rule.params]) {
        rulesToInsert[lastArea][rule.params] = []
      ***REMOVED***

      if ((!rule.hasDuplicates || !hasDuplicateName) && !rule.params) {
        // grid-template has no duplicates and not inside media rule

        getMSDecls(area, false, false)
          .reverse()
          .forEach(i =>
            gridAreaRule.prepend(
              Object.assign(i, {
                raws: {
                  between: gridArea.raws.between
                ***REMOVED***
              ***REMOVED***)
            )
          )

        rulesToInsert[lastArea].lastRule = gridAreaRule
        lastRuleIsSet = true
      ***REMOVED*** else if (rule.hasDuplicates && !rule.params && !selectorIsComplex) {
        // grid-template has duplicates and not inside media rule
        let cloned = gridAreaRule.clone()
        cloned.removeAll()

        getMSDecls(area, area.row.updateSpan, area.column.updateSpan)
          .reverse()
          .forEach(i =>
            cloned.prepend(
              Object.assign(i, {
                raws: {
                  between: gridArea.raws.between
                ***REMOVED***
              ***REMOVED***)
            )
          )

        cloned.selectors = changeDuplicateAreaSelectors(
          cloned.selectors,
          rule.selectors
        )

        if (rulesToInsert[lastArea].lastRule) {
          rulesToInsert[lastArea].lastRule.after(cloned)
        ***REMOVED***
        rulesToInsert[lastArea].lastRule = cloned
        lastRuleIsSet = true
      ***REMOVED*** else if (
        rule.hasDuplicates &&
        !rule.params &&
        selectorIsComplex &&
        gridAreaRule.selector.includes(rule.selectors[0])
      ) {
        // grid-template has duplicates and not inside media rule
        // and the selector is complex
        gridAreaRule.walkDecls(/-ms-grid-(row|column)/, d => d.remove())
        getMSDecls(area, area.row.updateSpan, area.column.updateSpan)
          .reverse()
          .forEach(i =>
            gridAreaRule.prepend(
              Object.assign(i, {
                raws: {
                  between: gridArea.raws.between
                ***REMOVED***
              ***REMOVED***)
            )
          )
      ***REMOVED*** else if (rule.params) {
        // grid-template is inside media rule
        // if we're inside media rule, we need to store prefixed rules
        // inside rulesToInsert object to be able to preserve the order of media
        // rules and merge them easily
        let cloned = gridAreaRule.clone()
        cloned.removeAll()

        getMSDecls(area, area.row.updateSpan, area.column.updateSpan)
          .reverse()
          .forEach(i =>
            cloned.prepend(
              Object.assign(i, {
                raws: {
                  between: gridArea.raws.between
                ***REMOVED***
              ***REMOVED***)
            )
          )

        if (rule.hasDuplicates && hasDuplicateName) {
          cloned.selectors = changeDuplicateAreaSelectors(
            cloned.selectors,
            rule.selectors
          )
        ***REMOVED***

        cloned.raws = rule.node.raws

        if (css.index(rule.node.parent) > gridAreaRuleIndex) {
          // append the prefixed rules right inside media rule
          // with grid-template
          rule.node.parent.append(cloned)
        ***REMOVED*** else {
          // store the rule to insert later
          rulesToInsert[lastArea][rule.params].push(cloned)
        ***REMOVED***

        // set new rule as last rule ONLY if we didn't set lastRule for
        // this grid-area before
        if (!lastRuleIsSet) {
          rulesToInsert[lastArea].lastRule = gridAreaMedia || gridAreaRule
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    return undefined
  ***REMOVED***)

  // append stored rules inside the media rules
  Object.keys(rulesToInsert).forEach(area => {
    let data = rulesToInsert[area]
    let lastRule = data.lastRule
    Object.keys(data)
      .reverse()
      .filter(p => p !== 'lastRule')
      .forEach(params => {
        if (data[params].length > 0 && lastRule) {
          lastRule.after({ name: 'media', params ***REMOVED***)
          lastRule.next().append(data[params])
        ***REMOVED***
      ***REMOVED***)
  ***REMOVED***)

  return undefined
***REMOVED***

/**
 * Warn user if grid area identifiers are not found
 * @param  {Object***REMOVED*** areas
 * @param  {Declaration***REMOVED*** decl
 * @param  {Result***REMOVED*** result
 * @return {void***REMOVED***
 */
exports.warnMissedAreas = warnMissedAreas

function warnMissedAreas(areas, decl, result) {
  let missed = Object.keys(areas)

  decl.root().walkDecls('grid-area', gridArea => {
    missed = missed.filter(e => e !== gridArea.value)
  ***REMOVED***)

  if (missed.length > 0) {
    decl.warn(result, 'Can not find grid areas: ' + missed.join(', '))
  ***REMOVED***

  return undefined
***REMOVED***

/**
 * compare selectors with grid-area rule and grid-template rule
 * show warning if grid-template selector is not found
 * (this function used for grid-area rule)
 * @param  {Declaration***REMOVED*** decl
 * @param  {Result***REMOVED*** result
 * @return {void***REMOVED***
 */
exports.warnTemplateSelectorNotFound = warnTemplateSelectorNotFound

function warnTemplateSelectorNotFound(decl, result) {
  let rule = decl.parent
  let root = decl.root()
  let duplicatesFound = false

  // slice selector array. Remove the last part (for comparison)
  let slicedSelectorArr = list
    .space(rule.selector)
    .filter(str => str !== '>')
    .slice(0, -1)

  // we need to compare only if selector is complex.
  // e.g '.grid-cell' is simple, but '.parent > .grid-cell' is complex
  if (slicedSelectorArr.length > 0) {
    let gridTemplateFound = false
    let foundAreaSelector = null

    root.walkDecls(/grid-template(-areas)?$/, d => {
      let parent = d.parent
      let templateSelectors = parent.selectors

      let { areas ***REMOVED*** = parseTemplate({ decl: d, gap: getGridGap(d) ***REMOVED***)
      let hasArea = areas[decl.value]

      // find the the matching selectors
      for (let tplSelector of templateSelectors) {
        if (gridTemplateFound) {
          break
        ***REMOVED***
        let tplSelectorArr = list.space(tplSelector).filter(str => str !== '>')

        gridTemplateFound = tplSelectorArr.every(
          (item, idx) => item === slicedSelectorArr[idx]
        )
      ***REMOVED***

      if (gridTemplateFound || !hasArea) {
        return true
      ***REMOVED***

      if (!foundAreaSelector) {
        foundAreaSelector = parent.selector
      ***REMOVED***

      // if we found the duplicate area with different selector
      if (foundAreaSelector && foundAreaSelector !== parent.selector) {
        duplicatesFound = true
      ***REMOVED***

      return undefined
    ***REMOVED***)

    // warn user if we didn't find template
    if (!gridTemplateFound && duplicatesFound) {
      decl.warn(
        result,
        'Autoprefixer cannot find a grid-template ' +
          `containing the duplicate grid-area "${decl.value***REMOVED***" ` +
          `with full selector matching: ${slicedSelectorArr.join(' ')***REMOVED***`
      )
    ***REMOVED***
  ***REMOVED***
***REMOVED***

/**
 * warn user if both grid-area and grid-(row|column)
 * declarations are present in the same rule
 * @param  {Declaration***REMOVED*** decl
 * @param  {Result***REMOVED*** result
 * @return {void***REMOVED***
 */
exports.warnIfGridRowColumnExists = warnIfGridRowColumnExists

function warnIfGridRowColumnExists(decl, result) {
  let rule = decl.parent
  let decls = []
  rule.walkDecls(/^grid-(row|column)/, d => {
    if (
      !d.prop.endsWith('-end') &&
      !d.value.startsWith('span') &&
      !d.prop.endsWith('-gap')
    ) {
      decls.push(d)
    ***REMOVED***
  ***REMOVED***)
  if (decls.length > 0) {
    decls.forEach(d => {
      d.warn(
        result,
        'You already have a grid-area declaration present in the rule. ' +
          `You should use either grid-area or ${d.prop***REMOVED***, not both`
      )
    ***REMOVED***)
  ***REMOVED***

  return undefined
***REMOVED***

// Gap utils

exports.getGridGap = getGridGap

function getGridGap(decl) {
  let gap = {***REMOVED***

  // try to find gap
  let testGap = /^(grid-)?((row|column)-)?gap$/
  decl.parent.walkDecls(testGap, ({ prop, value ***REMOVED***) => {
    if (/^(grid-)?gap$/.test(prop)) {
      let [row, , column] = parser(value).nodes

      gap.row = row && parser.stringify(row)
      gap.column = column ? parser.stringify(column) : gap.row
    ***REMOVED***
    if (/^(grid-)?row-gap$/.test(prop)) gap.row = value
    if (/^(grid-)?column-gap$/.test(prop)) gap.column = value
  ***REMOVED***)

  return gap
***REMOVED***

/**
 * parse media parameters (for example 'min-width: 500px')
 * @param  {String***REMOVED*** params parameter to parse
 * @return {***REMOVED***
 */
function parseMediaParams(params) {
  if (!params) {
    return []
  ***REMOVED***
  let parsed = parser(params)
  let prop
  let value

  parsed.walk(node => {
    if (node.type === 'word' && /min|max/g.test(node.value)) {
      prop = node.value
    ***REMOVED*** else if (node.value.includes('px')) {
      value = parseInt(node.value.replace(/\D/g, ''))
    ***REMOVED***
  ***REMOVED***)

  return [prop, value]
***REMOVED***

/**
 * Compare the selectors and decide if we
 * need to inherit gap from compared selector or not.
 * @type {String***REMOVED*** selA
 * @type {String***REMOVED*** selB
 * @return {Boolean***REMOVED***
 */
function shouldInheritGap(selA, selB) {
  let result

  // get arrays of selector split in 3-deep array
  let splitSelectorArrA = splitSelector(selA)
  let splitSelectorArrB = splitSelector(selB)

  if (splitSelectorArrA[0].length < splitSelectorArrB[0].length) {
    // abort if selectorA has lower descendant specificity then selectorB
    // (e.g '.grid' and '.hello .world .grid')
    return false
  ***REMOVED*** else if (splitSelectorArrA[0].length > splitSelectorArrB[0].length) {
    // if selectorA has higher descendant specificity then selectorB
    // (e.g '.foo .bar .grid' and '.grid')

    let idx = splitSelectorArrA[0].reduce((res, [item], index) => {
      let firstSelectorPart = splitSelectorArrB[0][0][0]
      if (item === firstSelectorPart) {
        return index
      ***REMOVED***
      return false
    ***REMOVED***, false)

    if (idx) {
      result = splitSelectorArrB[0].every((arr, index) => {
        return arr.every(
          (part, innerIndex) =>
            // because selectorA has more space elements, we need to slice
            // selectorA array by 'idx' number to compare them
            splitSelectorArrA[0].slice(idx)[index][innerIndex] === part
        )
      ***REMOVED***)
    ***REMOVED***
  ***REMOVED*** else {
    // if selectorA has the same descendant specificity as selectorB
    // this condition covers cases such as: '.grid.foo.bar' and '.grid'
    result = splitSelectorArrB.some(byCommaArr => {
      return byCommaArr.every((bySpaceArr, index) => {
        return bySpaceArr.every(
          (part, innerIndex) => splitSelectorArrA[0][index][innerIndex] === part
        )
      ***REMOVED***)
    ***REMOVED***)
  ***REMOVED***

  return result
***REMOVED***
/**
 * inherit grid gap values from the closest rule above
 * with the same selector
 * @param  {Declaration***REMOVED*** decl
 * @param  {Object***REMOVED*** gap gap values
 * @return {Object | Boolean***REMOVED*** return gap values or false (if not found)
 */
exports.inheritGridGap = inheritGridGap

function inheritGridGap(decl, gap) {
  let rule = decl.parent
  let mediaRule = getParentMedia(rule)
  let root = rule.root()

  // get an array of selector split in 3-deep array
  let splitSelectorArr = splitSelector(rule.selector)

  // abort if the rule already has gaps
  if (Object.keys(gap).length > 0) {
    return false
  ***REMOVED***

  // e.g ['min-width']
  let [prop] = parseMediaParams(mediaRule.params)

  let lastBySpace = splitSelectorArr[0]

  // get escaped value from the selector
  // if we have '.grid-2.foo.bar' selector, will be '\.grid\-2'
  let escaped = escapeRegexp(lastBySpace[lastBySpace.length - 1][0])

  let regexp = new RegExp(`(${escaped***REMOVED***$)|(${escaped***REMOVED***[,.])`)

  // find the closest rule with the same selector
  let closestRuleGap
  root.walkRules(regexp, r => {
    let gridGap

    // abort if are checking the same rule
    if (rule.toString() === r.toString()) {
      return false
    ***REMOVED***

    // find grid-gap values
    r.walkDecls('grid-gap', d => (gridGap = getGridGap(d)))

    // skip rule without gaps
    if (!gridGap || Object.keys(gridGap).length === 0) {
      return true
    ***REMOVED***

    // skip rules that should not be inherited from
    if (!shouldInheritGap(rule.selector, r.selector)) {
      return true
    ***REMOVED***

    let media = getParentMedia(r)
    if (media) {
      // if we are inside media, we need to check that media props match
      // e.g ('min-width' === 'min-width')
      let propToCompare = parseMediaParams(media.params)[0]
      if (propToCompare === prop) {
        closestRuleGap = gridGap
        return true
      ***REMOVED***
    ***REMOVED*** else {
      closestRuleGap = gridGap
      return true
    ***REMOVED***

    return undefined
  ***REMOVED***)

  // if we find the closest gap object
  if (closestRuleGap && Object.keys(closestRuleGap).length > 0) {
    return closestRuleGap
  ***REMOVED***
  return false
***REMOVED***

exports.warnGridGap = warnGridGap

function warnGridGap({ decl, gap, hasColumns, result ***REMOVED***) {
  let hasBothGaps = gap.row && gap.column
  if (!hasColumns && (hasBothGaps || (gap.column && !gap.row))) {
    delete gap.column
    decl.warn(
      result,
      'Can not implement grid-gap without grid-template-columns'
    )
  ***REMOVED***
***REMOVED***

/**
 * normalize the grid-template-rows/columns values
 * @param  {String***REMOVED*** str grid-template-rows/columns value
 * @return {Array***REMOVED*** normalized array with values
 * @example
 * let normalized = normalizeRowColumn('1fr repeat(2, 20px 50px) 1fr')
 * normalized // <= ['1fr', '20px', '50px', '20px', '50px', '1fr']
 */
function normalizeRowColumn(str) {
  let normalized = parser(str).nodes.reduce((result, node) => {
    if (node.type === 'function' && node.value === 'repeat') {
      let key = 'count'

      let [count, value] = node.nodes.reduce(
        (acc, n) => {
          if (n.type === 'word' && key === 'count') {
            acc[0] = Math.abs(parseInt(n.value))
            return acc
          ***REMOVED***
          if (n.type === 'div' && n.value === ',') {
            key = 'value'
            return acc
          ***REMOVED***
          if (key === 'value') {
            acc[1] += parser.stringify(n)
          ***REMOVED***
          return acc
        ***REMOVED***,
        [0, '']
      )

      if (count) {
        for (let i = 0; i < count; i++) {
          result.push(value)
        ***REMOVED***
      ***REMOVED***

      return result
    ***REMOVED***
    if (node.type === 'space') {
      return result
    ***REMOVED***
    return result.concat(parser.stringify(node))
  ***REMOVED***, [])

  return normalized
***REMOVED***

exports.autoplaceGridItems = autoplaceGridItems

/**
 * Autoplace grid items
 * @param {Declaration***REMOVED*** decl
 * @param {Result***REMOVED*** result
 * @param {Object***REMOVED*** gap gap values
 * @param {String***REMOVED*** autoflowValue grid-auto-flow value
 * @return {void***REMOVED***
 * @see https://github.com/postcss/autoprefixer/issues/1148
 */
function autoplaceGridItems(decl, result, gap, autoflowValue = 'row') {
  let { parent ***REMOVED*** = decl

  let rowDecl = parent.nodes.find(i => i.prop === 'grid-template-rows')
  let rows = normalizeRowColumn(rowDecl.value)
  let columns = normalizeRowColumn(decl.value)

  // Build array of area names with dummy values. If we have 3 columns and
  // 2 rows, filledRows will be equal to ['1 2 3', '4 5 6']
  let filledRows = rows.map((_, rowIndex) => {
    return Array.from(
      { length: columns.length ***REMOVED***,
      (v, k) => k + rowIndex * columns.length + 1
    ).join(' ')
  ***REMOVED***)

  let areas = parseGridAreas({ gap, rows: filledRows ***REMOVED***)
  let keys = Object.keys(areas)
  let items = keys.map(i => areas[i])

  // Change the order of cells if grid-auto-flow value is 'column'
  if (autoflowValue.includes('column')) {
    items = items.sort((a, b) => a.column.start - b.column.start)
  ***REMOVED***

  // Insert new rules
  items.reverse().forEach((item, index) => {
    let { column, row ***REMOVED*** = item
    let nodeSelector = parent.selectors
      .map(sel => sel + ` > *:nth-child(${keys.length - index***REMOVED***)`)
      .join(', ')

    // create new rule
    let node = parent.clone().removeAll()

    // change rule selector
    node.selector = nodeSelector

    // insert prefixed row/column values
    node.append({ prop: '-ms-grid-row', value: row.start ***REMOVED***)
    node.append({ prop: '-ms-grid-column', value: column.start ***REMOVED***)

    // insert rule
    parent.after(node)
  ***REMOVED***)

  return undefined
***REMOVED***
