// autoprefixer/lib/hacks/grid-template-areas.js
let Declaration = require('../declaration')
let {
  getGridGap,
  inheritGridGap,
  parseGridAreas,
  prefixTrackProp,
  prefixTrackValue,
  warnGridGap,
  warnMissedAreas
***REMOVED*** = require('./grid-utils')

function getGridRows(tpl) {
  return tpl
    .trim()
    .slice(1, -1)
    .split(/["']\s*["']?/g)
***REMOVED***

class GridTemplateAreas extends Declaration {
  /**
   * Translate grid-template-areas to separate -ms- prefixed properties
   */
  insert(decl, prefix, prefixes, result) {
    if (prefix !== '-ms-') return super.insert(decl, prefix, prefixes)

    let hasColumns = false
    let hasRows = false
    let parent = decl.parent
    let gap = getGridGap(decl)
    gap = inheritGridGap(decl, gap) || gap

    // remove already prefixed rows
    // to prevent doubling prefixes
    parent.walkDecls(/-ms-grid-rows/, i => i.remove())

    // add empty tracks to rows
    parent.walkDecls(/grid-template-(rows|columns)/, trackDecl => {
      if (trackDecl.prop === 'grid-template-rows') {
        hasRows = true
        let { prop, value ***REMOVED*** = trackDecl
        trackDecl.cloneBefore({
          prop: prefixTrackProp({ prefix, prop ***REMOVED***),
          value: prefixTrackValue({ gap: gap.row, value ***REMOVED***)
        ***REMOVED***)
      ***REMOVED*** else {
        hasColumns = true
      ***REMOVED***
    ***REMOVED***)

    let gridRows = getGridRows(decl.value)

    if (hasColumns && !hasRows && gap.row && gridRows.length > 1) {
      decl.cloneBefore({
        prop: '-ms-grid-rows',
        raws: {***REMOVED***,
        value: prefixTrackValue({
          gap: gap.row,
          value: `repeat(${gridRows.length***REMOVED***, auto)`
        ***REMOVED***)
      ***REMOVED***)
    ***REMOVED***

    // warnings
    warnGridGap({
      decl,
      gap,
      hasColumns,
      result
    ***REMOVED***)

    let areas = parseGridAreas({
      gap,
      rows: gridRows
    ***REMOVED***)

    warnMissedAreas(areas, decl, result)

    return decl
  ***REMOVED***
***REMOVED***

GridTemplateAreas.names = ['grid-template-areas']

module.exports = GridTemplateAreas
