// autoprefixer/lib/hacks/intrinsic.js
let OldValue = require('../old-value')
let Value = require('../value')

function regexp(name) {
  return new RegExp(`(^|[\\s,(])(${name***REMOVED***($|[\\s),]))`, 'gi')
***REMOVED***

class Intrinsic extends Value {
  add(decl, prefix) {
    if (decl.prop.includes('grid') && prefix !== '-webkit-') {
      return undefined
    ***REMOVED***
    return super.add(decl, prefix)
  ***REMOVED***

  isStretch() {
    return (
      this.name === 'stretch' ||
      this.name === 'fill' ||
      this.name === 'fill-available'
    )
  ***REMOVED***

  old(prefix) {
    let prefixed = prefix + this.name
    if (this.isStretch()) {
      if (prefix === '-moz-') {
        prefixed = '-moz-available'
      ***REMOVED*** else if (prefix === '-webkit-') {
        prefixed = '-webkit-fill-available'
      ***REMOVED***
    ***REMOVED***
    return new OldValue(this.name, prefixed, prefixed, regexp(prefixed))
  ***REMOVED***

  regexp() {
    if (!this.regexpCache) this.regexpCache = regexp(this.name)
    return this.regexpCache
  ***REMOVED***

  replace(string, prefix) {
    if (prefix === '-moz-' && this.isStretch()) {
      return string.replace(this.regexp(), '$1-moz-available$3')
    ***REMOVED***
    if (prefix === '-webkit-' && this.isStretch()) {
      return string.replace(this.regexp(), '$1-webkit-fill-available$3')
    ***REMOVED***
    return super.replace(string, prefix)
  ***REMOVED***
***REMOVED***

Intrinsic.names = [
  'max-content',
  'min-content',
  'fit-content',
  'fill',
  'fill-available',
  'stretch'
]

module.exports = Intrinsic
