// autoprefixer/lib/hacks/mask-composite.js
let Declaration = require('../declaration')

class MaskComposite extends Declaration {
  /**
   * Prefix mask-composite for webkit
   */
  insert(decl, prefix, prefixes) {
    let isCompositeProp = decl.prop === 'mask-composite'

    let compositeValues

    if (isCompositeProp) {
      compositeValues = decl.value.split(',')
    ***REMOVED*** else {
      compositeValues = decl.value.match(MaskComposite.regexp) || []
    ***REMOVED***

    compositeValues = compositeValues.map(el => el.trim()).filter(el => el)
    let hasCompositeValues = compositeValues.length

    let compositeDecl

    if (hasCompositeValues) {
      compositeDecl = this.clone(decl)
      compositeDecl.value = compositeValues
        .map(value => MaskComposite.oldValues[value] || value)
        .join(', ')

      if (compositeValues.includes('intersect')) {
        compositeDecl.value += ', xor'
      ***REMOVED***

      compositeDecl.prop = prefix + 'mask-composite'
    ***REMOVED***

    if (isCompositeProp) {
      if (!hasCompositeValues) {
        return undefined
      ***REMOVED***

      if (this.needCascade(decl)) {
        compositeDecl.raws.before = this.calcBefore(prefixes, decl, prefix)
      ***REMOVED***

      return decl.parent.insertBefore(decl, compositeDecl)
    ***REMOVED***

    let cloned = this.clone(decl)
    cloned.prop = prefix + cloned.prop

    if (hasCompositeValues) {
      cloned.value = cloned.value.replace(MaskComposite.regexp, '')
    ***REMOVED***

    if (this.needCascade(decl)) {
      cloned.raws.before = this.calcBefore(prefixes, decl, prefix)
    ***REMOVED***

    decl.parent.insertBefore(decl, cloned)

    if (!hasCompositeValues) {
      return decl
    ***REMOVED***

    if (this.needCascade(decl)) {
      compositeDecl.raws.before = this.calcBefore(prefixes, decl, prefix)
    ***REMOVED***
    return decl.parent.insertBefore(decl, compositeDecl)
  ***REMOVED***
***REMOVED***

MaskComposite.names = ['mask', 'mask-composite']

MaskComposite.oldValues = {
  add: 'source-over',
  exclude: 'xor',
  intersect: 'source-in',
  subtract: 'source-out'
***REMOVED***

MaskComposite.regexp = new RegExp(
  `\\s+(${Object.keys(MaskComposite.oldValues).join(
    '|'
  )***REMOVED***)\\b(?!\\))\\s*(?=[,])`,
  'ig'
)

module.exports = MaskComposite
