// hoopy/index.js
'use strict'

class Hoopy extends Array {
  constructor (size) {
    let index, isIndexOverflowed

    if (! isPositiveInteger(size)) {
      throw new TypeError('Argument `size` must be a positive integer.')
    ***REMOVED***

    super(size)

    this.grow = by => {
      if (! isPositiveInteger(by)) {
        throw new TypeError('Argument `by` must be a positive integer.')
      ***REMOVED***

      let i
      const newSize = size + by

      for (i = size; i < newSize; ++i) {
        this[i] = undefined
      ***REMOVED***

      if (isIndexOverflowed) {
        for (i = 0; i <= index; ++i) {
          let j = size + i
          if (j >= newSize) {
            j %= newSize
          ***REMOVED***
          this[j] = this[i]
          this[i] = undefined
        ***REMOVED***
      ***REMOVED***

      size = newSize
    ***REMOVED***

    return new Proxy(this, {
      get (target, key) {
        if (isInteger(key)) {
          return target[getIndex(key, size)]
        ***REMOVED***

        return target[key]
      ***REMOVED***,

      set (target, key, value) {
        if (isInteger(key)) {
          index = getIndex(key, size)
          target[index] = value

          if (Math.abs(key) >= size) {
            isIndexOverflowed = true
          ***REMOVED*** else {
            isIndexOverflowed = false
          ***REMOVED***
        ***REMOVED*** else {
          target[key] = value
        ***REMOVED***
        return true
      ***REMOVED***
    ***REMOVED***)
  ***REMOVED***
***REMOVED***

function isPositiveInteger (thing) {
  return isInteger(thing) && thing > 0
***REMOVED***

function isInteger (thing) {
  try {
    return +thing % 1 === 0
  ***REMOVED*** catch (error) {
    // Coercing symbols to numbers throws an error
  ***REMOVED***

  return false
***REMOVED***

function getIndex (key, size) {
  if (key === 0) {
    return 0
  ***REMOVED***

  if (key < 0) {
    return (size - Math.abs(key)) % size
  ***REMOVED***

  return key % size
***REMOVED***

function nop () {
  throw new Error('Not implemented')
***REMOVED***

Hoopy.prototype.push = nop
Hoopy.prototype.pop = nop
Hoopy.prototype.shift = nop
Hoopy.prototype.unshift = nop

module.exports = Hoopy
