// webpack/lib/util/LazySet.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const makeSerializable = require("./makeSerializable.js");

/**
 * @template T
 * @param {Set<T>***REMOVED*** targetSet set where items should be added
 * @param {Set<Iterable<T>>***REMOVED*** toMerge iterables to be merged
 * @returns {void***REMOVED***
 */
const merge = (targetSet, toMerge) => {
	for (const set of toMerge) {
		for (const item of set) {
			targetSet.add(item);
		***REMOVED***
	***REMOVED***
***REMOVED***;

/**
 * @template T
 * @param {Set<Iterable<T>>***REMOVED*** targetSet set where iterables should be added
 * @param {Array<LazySet<T>>***REMOVED*** toDeepMerge lazy sets to be flattened
 * @returns {void***REMOVED***
 */
const flatten = (targetSet, toDeepMerge) => {
	for (const set of toDeepMerge) {
		if (set._set.size > 0) targetSet.add(set._set);
		if (set._needMerge) {
			for (const mergedSet of set._toMerge) {
				targetSet.add(mergedSet);
			***REMOVED***
			flatten(targetSet, set._toDeepMerge);
		***REMOVED***
	***REMOVED***
***REMOVED***;

/**
 * Like Set but with an addAll method to eventually add items from another iterable.
 * Access methods make sure that all delayed operations are executed.
 * Iteration methods deopts to normal Set performance until clear is called again (because of the chance of modifications during iteration).
 * @template T
 */
class LazySet {
	/**
	 * @param {Iterable<T>=***REMOVED*** iterable init iterable
	 */
	constructor(iterable) {
		/** @type {Set<T>***REMOVED*** */
		this._set = new Set(iterable);
		/** @type {Set<Iterable<T>>***REMOVED*** */
		this._toMerge = new Set();
		/** @type {Array<LazySet<T>>***REMOVED*** */
		this._toDeepMerge = [];
		this._needMerge = false;
		this._deopt = false;
	***REMOVED***

	_flatten() {
		flatten(this._toMerge, this._toDeepMerge);
		this._toDeepMerge.length = 0;
	***REMOVED***

	_merge() {
		this._flatten();
		merge(this._set, this._toMerge);
		this._toMerge.clear();
		this._needMerge = false;
	***REMOVED***

	_isEmpty() {
		return (
			this._set.size === 0 &&
			this._toMerge.size === 0 &&
			this._toDeepMerge.length === 0
		);
	***REMOVED***

	get size() {
		if (this._needMerge) this._merge();
		return this._set.size;
	***REMOVED***

	/**
	 * @param {T***REMOVED*** item an item
	 * @returns {LazySet<T>***REMOVED*** itself
	 */
	add(item) {
		this._set.add(item);
		return this;
	***REMOVED***

	/**
	 * @param {Iterable<T> | LazySet<T>***REMOVED*** iterable a immutable iterable or another immutable LazySet which will eventually be merged into the Set
	 * @returns {LazySet<T>***REMOVED*** itself
	 */
	addAll(iterable) {
		if (this._deopt) {
			const _set = this._set;
			for (const item of iterable) {
				_set.add(item);
			***REMOVED***
		***REMOVED*** else {
			if (iterable instanceof LazySet) {
				if (iterable._isEmpty()) return this;
				this._toDeepMerge.push(iterable);
				this._needMerge = true;
				if (this._toDeepMerge.length > 100000) {
					this._flatten();
				***REMOVED***
			***REMOVED*** else {
				this._toMerge.add(iterable);
				this._needMerge = true;
			***REMOVED***
			if (this._toMerge.size > 100000) this._merge();
		***REMOVED***
		return this;
	***REMOVED***

	clear() {
		this._set.clear();
		this._toMerge.clear();
		this._toDeepMerge.length = 0;
		this._needMerge = false;
		this._deopt = false;
	***REMOVED***

	/**
	 * @param {T***REMOVED*** value an item
	 * @returns {boolean***REMOVED*** true, if the value was in the Set before
	 */
	delete(value) {
		if (this._needMerge) this._merge();
		return this._set.delete(value);
	***REMOVED***

	/**
	 * @returns {IterableIterator<[T, T]>***REMOVED*** entries
	 */
	entries() {
		this._deopt = true;
		if (this._needMerge) this._merge();
		return this._set.entries();
	***REMOVED***

	/**
	 * @template K
	 * @param {(value: T, value2: T, set: Set<T>) => void***REMOVED*** callbackFn function called for each entry
	 * @param {K***REMOVED*** thisArg this argument for the callbackFn
	 * @returns {void***REMOVED***
	 */
	forEach(callbackFn, thisArg) {
		this._deopt = true;
		if (this._needMerge) this._merge();
		// eslint-disable-next-line unicorn/no-array-for-each
		this._set.forEach(callbackFn, thisArg);
	***REMOVED***

	/**
	 * @param {T***REMOVED*** item an item
	 * @returns {boolean***REMOVED*** true, when the item is in the Set
	 */
	has(item) {
		if (this._needMerge) this._merge();
		return this._set.has(item);
	***REMOVED***

	/**
	 * @returns {IterableIterator<T>***REMOVED*** keys
	 */
	keys() {
		this._deopt = true;
		if (this._needMerge) this._merge();
		return this._set.keys();
	***REMOVED***

	/**
	 * @returns {IterableIterator<T>***REMOVED*** values
	 */
	values() {
		this._deopt = true;
		if (this._needMerge) this._merge();
		return this._set.values();
	***REMOVED***

	/**
	 * @returns {IterableIterator<T>***REMOVED*** iterable iterator
	 */
	[Symbol.iterator]() {
		this._deopt = true;
		if (this._needMerge) this._merge();
		return this._set[Symbol.iterator]();
	***REMOVED***

	/* istanbul ignore next */
	get [Symbol.toStringTag]() {
		return "LazySet";
	***REMOVED***

	/**
	 * @param {import("../serialization/ObjectMiddleware").ObjectSerializerContext***REMOVED*** context context
	 */
	serialize({ write ***REMOVED***) {
		if (this._needMerge) this._merge();
		write(this._set.size);
		for (const item of this._set) write(item);
	***REMOVED***

	/**
	 * @template T
	 * @param {import("../serialization/ObjectMiddleware").ObjectDeserializerContext***REMOVED*** context context
	 * @returns {LazySet<T>***REMOVED*** lazy set
	 */
	static deserialize({ read ***REMOVED***) {
		const count = read();
		const items = [];
		for (let i = 0; i < count; i++) {
			items.push(read());
		***REMOVED***
		return new LazySet(items);
	***REMOVED***
***REMOVED***

makeSerializable(LazySet, "webpack/lib/util/LazySet");

module.exports = LazySet;
