// webpack/lib/util/LazyBucketSortedSet.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const { first ***REMOVED*** = require("./SetHelpers");
const SortableSet = require("./SortableSet");

/**
 * @template T
 * @template K
 * @typedef {(item: T) => K***REMOVED*** GetKey
 */

/**
 * @template T
 * @typedef {(a: T, n: T) => number***REMOVED*** Comparator
 */

/**
 * @template T
 * @template K
 * @typedef {LazyBucketSortedSet<T, K> | SortableSet<T>***REMOVED*** Entry
 */

/**
 * @template T
 * @template K
 * @typedef {GetKey<T, K> | Comparator<K> | Comparator<T>***REMOVED*** Arg
 */

/**
 * Multi layer bucket sorted set:
 * Supports adding non-existing items (DO NOT ADD ITEM TWICE),
 * Supports removing exiting items (DO NOT REMOVE ITEM NOT IN SET),
 * Supports popping the first items according to defined order,
 * Supports iterating all items without order,
 * Supports updating an item in an efficient way,
 * Supports size property, which is the number of items,
 * Items are lazy partially sorted when needed
 * @template T
 * @template K
 */
class LazyBucketSortedSet {
	/**
	 * @param {GetKey<T, K>***REMOVED*** getKey function to get key from item
	 * @param {Comparator<K>=***REMOVED*** comparator comparator to sort keys
	 * @param {...Arg<T, K>***REMOVED*** args more pairs of getKey and comparator plus optional final comparator for the last layer
	 */
	constructor(getKey, comparator, ...args) {
		this._getKey = getKey;
		this._innerArgs = args;
		this._leaf = args.length <= 1;
		this._keys = new SortableSet(undefined, comparator);
		/** @type {Map<K, Entry<T, K>>***REMOVED*** */
		this._map = new Map();
		this._unsortedItems = new Set();
		this.size = 0;
	***REMOVED***

	/**
	 * @param {T***REMOVED*** item an item
	 * @returns {void***REMOVED***
	 */
	add(item) {
		this.size++;
		this._unsortedItems.add(item);
	***REMOVED***

	/**
	 * @param {K***REMOVED*** key key of item
	 * @param {T***REMOVED*** item the item
	 * @returns {void***REMOVED***
	 */
	_addInternal(key, item) {
		let entry = this._map.get(key);
		if (entry === undefined) {
			entry = this._leaf
				? new SortableSet(
						undefined,
						/** @type {Comparator<T>***REMOVED*** */
						(this._innerArgs[0])
					)
				: new LazyBucketSortedSet(
						.../** @type {[GetKey<T, K>, Comparator<K>]***REMOVED*** */
						(this._innerArgs)
					);
			this._keys.add(key);
			this._map.set(key, entry);
		***REMOVED***
		entry.add(item);
	***REMOVED***

	/**
	 * @param {T***REMOVED*** item an item
	 * @returns {void***REMOVED***
	 */
	delete(item) {
		this.size--;
		if (this._unsortedItems.has(item)) {
			this._unsortedItems.delete(item);
			return;
		***REMOVED***
		const key = this._getKey(item);
		const entry = /** @type {Entry<T, K>***REMOVED*** */ (this._map.get(key));
		entry.delete(item);
		if (entry.size === 0) {
			this._deleteKey(key);
		***REMOVED***
	***REMOVED***

	/**
	 * @param {K***REMOVED*** key key to be removed
	 * @returns {void***REMOVED***
	 */
	_deleteKey(key) {
		this._keys.delete(key);
		this._map.delete(key);
	***REMOVED***

	/**
	 * @returns {T | undefined***REMOVED*** an item
	 */
	popFirst() {
		if (this.size === 0) return;
		this.size--;
		if (this._unsortedItems.size > 0) {
			for (const item of this._unsortedItems) {
				const key = this._getKey(item);
				this._addInternal(key, item);
			***REMOVED***
			this._unsortedItems.clear();
		***REMOVED***
		this._keys.sort();
		const key = /** @type {K***REMOVED*** */ (first(this._keys));
		const entry = this._map.get(key);
		if (this._leaf) {
			const leafEntry = /** @type {SortableSet<T>***REMOVED*** */ (entry);
			leafEntry.sort();
			const item = /** @type {T***REMOVED*** */ (first(leafEntry));
			leafEntry.delete(item);
			if (leafEntry.size === 0) {
				this._deleteKey(key);
			***REMOVED***
			return item;
		***REMOVED***
		const nodeEntry =
			/** @type {LazyBucketSortedSet<T, K>***REMOVED*** */
			(entry);
		const item = nodeEntry.popFirst();
		if (nodeEntry.size === 0) {
			this._deleteKey(key);
		***REMOVED***
		return item;
	***REMOVED***

	/**
	 * @param {T***REMOVED*** item to be updated item
	 * @returns {(remove?: true) => void***REMOVED*** finish update
	 */
	startUpdate(item) {
		if (this._unsortedItems.has(item)) {
			return remove => {
				if (remove) {
					this._unsortedItems.delete(item);
					this.size--;
				***REMOVED***
			***REMOVED***;
		***REMOVED***
		const key = this._getKey(item);
		if (this._leaf) {
			const oldEntry = /** @type {SortableSet<T>***REMOVED*** */ (this._map.get(key));
			return remove => {
				if (remove) {
					this.size--;
					oldEntry.delete(item);
					if (oldEntry.size === 0) {
						this._deleteKey(key);
					***REMOVED***
					return;
				***REMOVED***
				const newKey = this._getKey(item);
				if (key === newKey) {
					// This flags the sortable set as unordered
					oldEntry.add(item);
				***REMOVED*** else {
					oldEntry.delete(item);
					if (oldEntry.size === 0) {
						this._deleteKey(key);
					***REMOVED***
					this._addInternal(newKey, item);
				***REMOVED***
			***REMOVED***;
		***REMOVED***
		const oldEntry =
			/** @type {LazyBucketSortedSet<T, K>***REMOVED*** */
			(this._map.get(key));
		const finishUpdate = oldEntry.startUpdate(item);
		return remove => {
			if (remove) {
				this.size--;
				finishUpdate(true);
				if (oldEntry.size === 0) {
					this._deleteKey(key);
				***REMOVED***
				return;
			***REMOVED***
			const newKey = this._getKey(item);
			if (key === newKey) {
				finishUpdate();
			***REMOVED*** else {
				finishUpdate(true);
				if (oldEntry.size === 0) {
					this._deleteKey(key);
				***REMOVED***
				this._addInternal(newKey, item);
			***REMOVED***
		***REMOVED***;
	***REMOVED***

	/**
	 * @param {Iterator<T>[]***REMOVED*** iterators list of iterators to append to
	 * @returns {void***REMOVED***
	 */
	_appendIterators(iterators) {
		if (this._unsortedItems.size > 0)
			iterators.push(this._unsortedItems[Symbol.iterator]());
		for (const key of this._keys) {
			const entry = this._map.get(key);
			if (this._leaf) {
				const leafEntry = /** @type {SortableSet<T>***REMOVED*** */ (entry);
				const iterator = leafEntry[Symbol.iterator]();
				iterators.push(iterator);
			***REMOVED*** else {
				const nodeEntry =
					/** @type {LazyBucketSortedSet<T, K>***REMOVED*** */
					(entry);
				nodeEntry._appendIterators(iterators);
			***REMOVED***
		***REMOVED***
	***REMOVED***

	/**
	 * @returns {Iterator<T>***REMOVED*** the iterator
	 */
	[Symbol.iterator]() {
		/** @type {Iterator<T>[]***REMOVED*** */
		const iterators = [];
		this._appendIterators(iterators);
		iterators.reverse();
		let currentIterator =
			/** @type {Iterator<T>***REMOVED*** */
			(iterators.pop());
		return {
			next: () => {
				const res = currentIterator.next();
				if (res.done) {
					if (iterators.length === 0) return res;
					currentIterator = /** @type {Iterator<T>***REMOVED*** */ (iterators.pop());
					return currentIterator.next();
				***REMOVED***
				return res;
			***REMOVED***
		***REMOVED***;
	***REMOVED***
***REMOVED***

module.exports = LazyBucketSortedSet;
