// webpack/lib/util/WeakTupleMap.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

/**
 * @template {EXPECTED_ANY[]***REMOVED*** T
 * @template V
 * @typedef {Map<EXPECTED_ANY, WeakTupleMap<T, V>>***REMOVED*** M
 */

/**
 * @template {EXPECTED_ANY[]***REMOVED*** T
 * @template V
 * @typedef {WeakMap<EXPECTED_OBJECT, WeakTupleMap<T, V>>***REMOVED*** W
 */

/**
 * @param {EXPECTED_ANY***REMOVED*** thing thing
 * @returns {boolean***REMOVED*** true if is weak
 */
const isWeakKey = thing => typeof thing === "object" && thing !== null;

/**
 * @template {unknown[]***REMOVED*** T
 * @typedef {T extends readonly (infer ElementType)[] ? ElementType : never***REMOVED*** ArrayElement
 */

/**
 * @template {EXPECTED_ANY[]***REMOVED*** K
 * @template V
 */
class WeakTupleMap {
	constructor() {
		/** @private */
		this.f = 0;
		/**
		 * @private
		 * @type {V | undefined***REMOVED***
		 */
		this.v = undefined;
		/**
		 * @private
		 * @type {M<K, V> | undefined***REMOVED***
		 */
		this.m = undefined;
		/**
		 * @private
		 * @type {W<K, V> | undefined***REMOVED***
		 */
		this.w = undefined;
	***REMOVED***

	/**
	 * @param {[...K, V]***REMOVED*** args tuple
	 * @returns {void***REMOVED***
	 */
	set(...args) {
		/** @type {WeakTupleMap<K, V>***REMOVED*** */
		let node = this;
		for (let i = 0; i < args.length - 1; i++) {
			node = node._get(/** @type {ArrayElement<K>***REMOVED*** */ (args[i]));
		***REMOVED***
		node._setValue(/** @type {V***REMOVED*** */ (args[args.length - 1]));
	***REMOVED***

	/**
	 * @param {K***REMOVED*** args tuple
	 * @returns {boolean***REMOVED*** true, if the tuple is in the Set
	 */
	has(...args) {
		/** @type {WeakTupleMap<K, V> | undefined***REMOVED*** */
		let node = this;
		for (let i = 0; i < args.length; i++) {
			node = node._peek(/** @type {ArrayElement<K>***REMOVED*** */ (args[i]));
			if (node === undefined) return false;
		***REMOVED***
		return node._hasValue();
	***REMOVED***

	/**
	 * @param {K***REMOVED*** args tuple
	 * @returns {V | undefined***REMOVED*** the value
	 */
	get(...args) {
		/** @type {WeakTupleMap<K, V> | undefined***REMOVED*** */
		let node = this;
		for (let i = 0; i < args.length; i++) {
			node = node._peek(/** @type {ArrayElement<K>***REMOVED*** */ (args[i]));
			if (node === undefined) return;
		***REMOVED***
		return node._getValue();
	***REMOVED***

	/**
	 * @param {[...K, (...args: K) => V]***REMOVED*** args tuple
	 * @returns {V***REMOVED*** the value
	 */
	provide(...args) {
		/** @type {WeakTupleMap<K, V>***REMOVED*** */
		let node = this;
		for (let i = 0; i < args.length - 1; i++) {
			node = node._get(/** @type {ArrayElement<K>***REMOVED*** */ (args[i]));
		***REMOVED***
		if (node._hasValue()) return /** @type {V***REMOVED*** */ (node._getValue());
		const fn = /** @type {(...args: K) => V***REMOVED*** */ (args[args.length - 1]);
		const newValue = fn(.../** @type {K***REMOVED*** */ (args.slice(0, -1)));
		node._setValue(newValue);
		return newValue;
	***REMOVED***

	/**
	 * @param {K***REMOVED*** args tuple
	 * @returns {void***REMOVED***
	 */
	delete(...args) {
		/** @type {WeakTupleMap<K, V> | undefined***REMOVED*** */
		let node = this;
		for (let i = 0; i < args.length; i++) {
			node = node._peek(/** @type {ArrayElement<K>***REMOVED*** */ (args[i]));
			if (node === undefined) return;
		***REMOVED***
		node._deleteValue();
	***REMOVED***

	/**
	 * @returns {void***REMOVED***
	 */
	clear() {
		this.f = 0;
		this.v = undefined;
		this.w = undefined;
		this.m = undefined;
	***REMOVED***

	_getValue() {
		return this.v;
	***REMOVED***

	_hasValue() {
		return (this.f & 1) === 1;
	***REMOVED***

	/**
	 * @param {V***REMOVED*** v value
	 * @private
	 */
	_setValue(v) {
		this.f |= 1;
		this.v = v;
	***REMOVED***

	_deleteValue() {
		this.f &= 6;
		this.v = undefined;
	***REMOVED***

	/**
	 * @param {ArrayElement<K>***REMOVED*** thing thing
	 * @returns {WeakTupleMap<K, V> | undefined***REMOVED*** thing
	 * @private
	 */
	_peek(thing) {
		if (isWeakKey(thing)) {
			if ((this.f & 4) !== 4) return;
			return /** @type {WeakMap<ArrayElement<K>, WeakTupleMap<K, V>>***REMOVED*** */ (
				this.w
			).get(thing);
		***REMOVED***
		if ((this.f & 2) !== 2) return;
		return /** @type {Map<ArrayElement<K>, WeakTupleMap<K, V>>***REMOVED*** */ (this.m).get(
			thing
		);
	***REMOVED***

	/**
	 * @private
	 * @param {ArrayElement<K>***REMOVED*** thing thing
	 * @returns {WeakTupleMap<K, V>***REMOVED*** value
	 */
	_get(thing) {
		if (isWeakKey(thing)) {
			if ((this.f & 4) !== 4) {
				/** @type {W<K, V>***REMOVED*** */
				const newMap = new WeakMap();
				this.f |= 4;
				/** @type {WeakTupleMap<K, V>***REMOVED*** */
				const newNode = new WeakTupleMap();
				(this.w = newMap).set(thing, newNode);
				return newNode;
			***REMOVED***
			const entry = /** @type {W<K, V>***REMOVED*** */ (this.w).get(thing);
			if (entry !== undefined) {
				return entry;
			***REMOVED***
			/** @type {WeakTupleMap<K, V>***REMOVED*** */
			const newNode = new WeakTupleMap();
			/** @type {W<K, V>***REMOVED*** */
			(this.w).set(thing, newNode);
			return newNode;
		***REMOVED***
		if ((this.f & 2) !== 2) {
			/** @type {M<K, V>***REMOVED*** */
			const newMap = new Map();
			this.f |= 2;
			/** @type {WeakTupleMap<K, V>***REMOVED*** */
			const newNode = new WeakTupleMap();
			(this.m = newMap).set(thing, newNode);
			return newNode;
		***REMOVED***
		const entry =
			/** @type {M<K, V>***REMOVED*** */
			(this.m).get(thing);
		if (entry !== undefined) {
			return entry;
		***REMOVED***
		/** @type {WeakTupleMap<K, V>***REMOVED*** */
		const newNode = new WeakTupleMap();
		/** @type {M<K, V>***REMOVED*** */
		(this.m).set(thing, newNode);
		return newNode;
	***REMOVED***
***REMOVED***

module.exports = WeakTupleMap;
