// tailwindcss/src/util/pluginUtils.js
import escapeCommas from './escapeCommas'
import { withAlphaValue ***REMOVED*** from './withAlphaVariable'
import {
  normalize,
  length,
  number,
  percentage,
  url,
  color as validateColor,
  genericName,
  familyName,
  image,
  absoluteSize,
  relativeSize,
  position,
  lineWidth,
  shadow,
***REMOVED*** from './dataTypes'
import negateValue from './negateValue'
import { backgroundSize ***REMOVED*** from './validateFormalSyntax'
import { flagEnabled ***REMOVED*** from '../featureFlags.js'

/**
 * @param {import('postcss-selector-parser').Container***REMOVED*** selectors
 * @param {(className: string) => string***REMOVED*** updateClass
 * @returns {string***REMOVED***
 */
export function updateAllClasses(selectors, updateClass) {
  selectors.walkClasses((sel) => {
    sel.value = updateClass(sel.value)

    if (sel.raws && sel.raws.value) {
      sel.raws.value = escapeCommas(sel.raws.value)
    ***REMOVED***
  ***REMOVED***)
***REMOVED***

function resolveArbitraryValue(modifier, validate) {
  if (!isArbitraryValue(modifier)) {
    return undefined
  ***REMOVED***

  let value = modifier.slice(1, -1)

  if (!validate(value)) {
    return undefined
  ***REMOVED***

  return normalize(value)
***REMOVED***

function asNegativeValue(modifier, lookup = {***REMOVED***, validate) {
  let positiveValue = lookup[modifier]

  if (positiveValue !== undefined) {
    return negateValue(positiveValue)
  ***REMOVED***

  if (isArbitraryValue(modifier)) {
    let resolved = resolveArbitraryValue(modifier, validate)

    if (resolved === undefined) {
      return undefined
    ***REMOVED***

    return negateValue(resolved)
  ***REMOVED***
***REMOVED***

export function asValue(modifier, options = {***REMOVED***, { validate = () => true ***REMOVED*** = {***REMOVED***) {
  let value = options.values?.[modifier]

  if (value !== undefined) {
    return value
  ***REMOVED***

  if (options.supportsNegativeValues && modifier.startsWith('-')) {
    return asNegativeValue(modifier.slice(1), options.values, validate)
  ***REMOVED***

  return resolveArbitraryValue(modifier, validate)
***REMOVED***

function isArbitraryValue(input) {
  return input.startsWith('[') && input.endsWith(']')
***REMOVED***

function splitUtilityModifier(modifier) {
  let slashIdx = modifier.lastIndexOf('/')

  // If the `/` is inside an arbitrary, we want to find the previous one if any
  // This logic probably isn't perfect but it should work for most cases
  let arbitraryStartIdx = modifier.lastIndexOf('[', slashIdx)
  let arbitraryEndIdx = modifier.indexOf(']', slashIdx)

  let isNextToArbitrary = modifier[slashIdx - 1] === ']' || modifier[slashIdx + 1] === '['

  // Backtrack to the previous `/` if the one we found was inside an arbitrary
  if (!isNextToArbitrary) {
    if (arbitraryStartIdx !== -1 && arbitraryEndIdx !== -1) {
      if (arbitraryStartIdx < slashIdx && slashIdx < arbitraryEndIdx) {
        slashIdx = modifier.lastIndexOf('/', arbitraryStartIdx)
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  if (slashIdx === -1 || slashIdx === modifier.length - 1) {
    return [modifier, undefined]
  ***REMOVED***

  let arbitrary = isArbitraryValue(modifier)

  // The modifier could be of the form `[foo]/[bar]`
  // We want to handle this case properly
  // without affecting `[foo/bar]`
  if (arbitrary && !modifier.includes(']/[')) {
    return [modifier, undefined]
  ***REMOVED***

  return [modifier.slice(0, slashIdx), modifier.slice(slashIdx + 1)]
***REMOVED***

export function parseColorFormat(value) {
  if (typeof value === 'string' && value.includes('<alpha-value>')) {
    let oldValue = value

    return ({ opacityValue = 1 ***REMOVED***) => oldValue.replace(/<alpha-value>/g, opacityValue)
  ***REMOVED***

  return value
***REMOVED***

function unwrapArbitraryModifier(modifier) {
  return normalize(modifier.slice(1, -1))
***REMOVED***

export function asColor(modifier, options = {***REMOVED***, { tailwindConfig = {***REMOVED*** ***REMOVED*** = {***REMOVED***) {
  if (options.values?.[modifier] !== undefined) {
    return parseColorFormat(options.values?.[modifier])
  ***REMOVED***

  // TODO: Hoist this up to getMatchingTypes or something
  // We do this here because we need the alpha value (if any)
  let [color, alpha] = splitUtilityModifier(modifier)

  if (alpha !== undefined) {
    let normalizedColor =
      options.values?.[color] ?? (isArbitraryValue(color) ? color.slice(1, -1) : undefined)

    if (normalizedColor === undefined) {
      return undefined
    ***REMOVED***

    normalizedColor = parseColorFormat(normalizedColor)

    if (isArbitraryValue(alpha)) {
      return withAlphaValue(normalizedColor, unwrapArbitraryModifier(alpha))
    ***REMOVED***

    if (tailwindConfig.theme?.opacity?.[alpha] === undefined) {
      return undefined
    ***REMOVED***

    return withAlphaValue(normalizedColor, tailwindConfig.theme.opacity[alpha])
  ***REMOVED***

  return asValue(modifier, options, { validate: validateColor ***REMOVED***)
***REMOVED***

export function asLookupValue(modifier, options = {***REMOVED***) {
  return options.values?.[modifier]
***REMOVED***

function guess(validate) {
  return (modifier, options) => {
    return asValue(modifier, options, { validate ***REMOVED***)
  ***REMOVED***
***REMOVED***

export let typeMap = {
  any: asValue,
  color: asColor,
  url: guess(url),
  image: guess(image),
  length: guess(length),
  percentage: guess(percentage),
  position: guess(position),
  lookup: asLookupValue,
  'generic-name': guess(genericName),
  'family-name': guess(familyName),
  number: guess(number),
  'line-width': guess(lineWidth),
  'absolute-size': guess(absoluteSize),
  'relative-size': guess(relativeSize),
  shadow: guess(shadow),
  size: guess(backgroundSize),
***REMOVED***

let supportedTypes = Object.keys(typeMap)

function splitAtFirst(input, delim) {
  let idx = input.indexOf(delim)
  if (idx === -1) return [undefined, input]
  return [input.slice(0, idx), input.slice(idx + 1)]
***REMOVED***

export function coerceValue(types, modifier, options, tailwindConfig) {
  if (options.values && modifier in options.values) {
    for (let { type ***REMOVED*** of types ?? []) {
      let result = typeMap[type](modifier, options, {
        tailwindConfig,
      ***REMOVED***)

      if (result === undefined) {
        continue
      ***REMOVED***

      return [result, type, null]
    ***REMOVED***
  ***REMOVED***

  if (isArbitraryValue(modifier)) {
    let arbitraryValue = modifier.slice(1, -1)
    let [explicitType, value] = splitAtFirst(arbitraryValue, ':')

    // It could be that this resolves to `url(https` which is not a valid
    // identifier. We currently only support "simple" words with dashes or
    // underscores. E.g.: family-name
    if (!/^[\w-_]+$/g.test(explicitType)) {
      value = arbitraryValue
    ***REMOVED***

    //
    else if (explicitType !== undefined && !supportedTypes.includes(explicitType)) {
      return []
    ***REMOVED***

    if (value.length > 0 && supportedTypes.includes(explicitType)) {
      return [asValue(`[${value***REMOVED***]`, options), explicitType, null]
    ***REMOVED***
  ***REMOVED***

  let matches = getMatchingTypes(types, modifier, options, tailwindConfig)

  // Find first matching type
  for (let match of matches) {
    return match
  ***REMOVED***

  return []
***REMOVED***

/**
 *
 * @param {{type: string***REMOVED***[]***REMOVED*** types
 * @param {string***REMOVED*** rawModifier
 * @param {any***REMOVED*** options
 * @param {any***REMOVED*** tailwindConfig
 * @returns {Iterator<[value: string, type: string, modifier: string | null]>***REMOVED***
 */
export function* getMatchingTypes(types, rawModifier, options, tailwindConfig) {
  let modifiersEnabled = flagEnabled(tailwindConfig, 'generalizedModifiers')

  let [modifier, utilityModifier] = splitUtilityModifier(rawModifier)

  let canUseUtilityModifier =
    modifiersEnabled &&
    options.modifiers != null &&
    (options.modifiers === 'any' ||
      (typeof options.modifiers === 'object' &&
        ((utilityModifier && isArbitraryValue(utilityModifier)) ||
          utilityModifier in options.modifiers)))

  if (!canUseUtilityModifier) {
    modifier = rawModifier
    utilityModifier = undefined
  ***REMOVED***

  if (utilityModifier !== undefined && modifier === '') {
    modifier = 'DEFAULT'
  ***REMOVED***

  // Check the full value first
  // TODO: Move to asValue… somehow
  if (utilityModifier !== undefined) {
    if (typeof options.modifiers === 'object') {
      let configValue = options.modifiers?.[utilityModifier] ?? null
      if (configValue !== null) {
        utilityModifier = configValue
      ***REMOVED*** else if (isArbitraryValue(utilityModifier)) {
        utilityModifier = unwrapArbitraryModifier(utilityModifier)
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  for (let { type ***REMOVED*** of types ?? []) {
    let result = typeMap[type](modifier, options, {
      tailwindConfig,
    ***REMOVED***)

    if (result === undefined) {
      continue
    ***REMOVED***

    yield [result, type, utilityModifier ?? null]
  ***REMOVED***
***REMOVED***
