// tailwindcss/src/util/resolveConfig.js
import negateValue from './negateValue'
import corePluginList from '../corePluginList'
import configurePlugins from './configurePlugins'
import colors from '../public/colors'
import { defaults ***REMOVED*** from './defaults'
import { toPath ***REMOVED*** from './toPath'
import { normalizeConfig ***REMOVED*** from './normalizeConfig'
import isPlainObject from './isPlainObject'
import { cloneDeep ***REMOVED*** from './cloneDeep'
import { parseColorFormat ***REMOVED*** from './pluginUtils'
import { withAlphaValue ***REMOVED*** from './withAlphaVariable'
import toColorValue from './toColorValue'

function isFunction(input) {
  return typeof input === 'function'
***REMOVED***

function mergeWith(target, ...sources) {
  let customizer = sources.pop()

  for (let source of sources) {
    for (let k in source) {
      let merged = customizer(target[k], source[k])

      if (merged === undefined) {
        if (isPlainObject(target[k]) && isPlainObject(source[k])) {
          target[k] = mergeWith({***REMOVED***, target[k], source[k], customizer)
        ***REMOVED*** else {
          target[k] = source[k]
        ***REMOVED***
      ***REMOVED*** else {
        target[k] = merged
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  return target
***REMOVED***

const configUtils = {
  colors,
  negative(scale) {
    // TODO: Log that this function isn't really needed anymore?
    return Object.keys(scale)
      .filter((key) => scale[key] !== '0')
      .reduce((negativeScale, key) => {
        let negativeValue = negateValue(scale[key])

        if (negativeValue !== undefined) {
          negativeScale[`-${key***REMOVED***`] = negativeValue
        ***REMOVED***

        return negativeScale
      ***REMOVED***, {***REMOVED***)
  ***REMOVED***,
  breakpoints(screens) {
    return Object.keys(screens)
      .filter((key) => typeof screens[key] === 'string')
      .reduce(
        (breakpoints, key) => ({
          ...breakpoints,
          [`screen-${key***REMOVED***`]: screens[key],
        ***REMOVED***),
        {***REMOVED***
      )
  ***REMOVED***,
***REMOVED***

function value(valueToResolve, ...args) {
  return isFunction(valueToResolve) ? valueToResolve(...args) : valueToResolve
***REMOVED***

function collectExtends(items) {
  return items.reduce((merged, { extend ***REMOVED***) => {
    return mergeWith(merged, extend, (mergedValue, extendValue) => {
      if (mergedValue === undefined) {
        return [extendValue]
      ***REMOVED***

      if (Array.isArray(mergedValue)) {
        return [extendValue, ...mergedValue]
      ***REMOVED***

      return [extendValue, mergedValue]
    ***REMOVED***)
  ***REMOVED***, {***REMOVED***)
***REMOVED***

function mergeThemes(themes) {
  return {
    ...themes.reduce((merged, theme) => defaults(merged, theme), {***REMOVED***),

    // In order to resolve n config objects, we combine all of their `extend` properties
    // into arrays instead of objects so they aren't overridden.
    extend: collectExtends(themes),
  ***REMOVED***
***REMOVED***

function mergeExtensionCustomizer(merged, value) {
  // When we have an array of objects, we do want to merge it
  if (Array.isArray(merged) && isPlainObject(merged[0])) {
    return merged.concat(value)
  ***REMOVED***

  // When the incoming value is an array, and the existing config is an object, prepend the existing object
  if (Array.isArray(value) && isPlainObject(value[0]) && isPlainObject(merged)) {
    return [merged, ...value]
  ***REMOVED***

  // Override arrays (for example for font-families, box-shadows, ...)
  if (Array.isArray(value)) {
    return value
  ***REMOVED***

  // Execute default behaviour
  return undefined
***REMOVED***

function mergeExtensions({ extend, ...theme ***REMOVED***) {
  return mergeWith(theme, extend, (themeValue, extensions) => {
    // The `extend` property is an array, so we need to check if it contains any functions
    if (!isFunction(themeValue) && !extensions.some(isFunction)) {
      return mergeWith({***REMOVED***, themeValue, ...extensions, mergeExtensionCustomizer)
    ***REMOVED***

    return (resolveThemePath, utils) =>
      mergeWith(
        {***REMOVED***,
        ...[themeValue, ...extensions].map((e) => value(e, resolveThemePath, utils)),
        mergeExtensionCustomizer
      )
  ***REMOVED***)
***REMOVED***

/**
 *
 * @param {string***REMOVED*** key
 * @return {Iterable<string[] & {alpha: string | undefined***REMOVED***>***REMOVED***
 */
function* toPaths(key) {
  let path = toPath(key)

  if (path.length === 0) {
    return
  ***REMOVED***

  yield path

  if (Array.isArray(key)) {
    return
  ***REMOVED***

  let pattern = /^(.*?)\s*\/\s*([^/]+)$/
  let matches = key.match(pattern)

  if (matches !== null) {
    let [, prefix, alpha] = matches

    let newPath = toPath(prefix)
    newPath.alpha = alpha

    yield newPath
  ***REMOVED***
***REMOVED***

function resolveFunctionKeys(object) {
  // theme('colors.red.500 / 0.5') -> ['colors', 'red', '500 / 0', '5]

  const resolvePath = (key, defaultValue) => {
    for (const path of toPaths(key)) {
      let index = 0
      let val = object

      while (val !== undefined && val !== null && index < path.length) {
        val = val[path[index++]]

        let shouldResolveAsFn =
          isFunction(val) && (path.alpha === undefined || index <= path.length - 1)

        val = shouldResolveAsFn ? val(resolvePath, configUtils) : val
      ***REMOVED***

      if (val !== undefined) {
        if (path.alpha !== undefined) {
          let normalized = parseColorFormat(val)

          return withAlphaValue(normalized, path.alpha, toColorValue(normalized))
        ***REMOVED***

        if (isPlainObject(val)) {
          return cloneDeep(val)
        ***REMOVED***

        return val
      ***REMOVED***
    ***REMOVED***

    return defaultValue
  ***REMOVED***

  Object.assign(resolvePath, {
    theme: resolvePath,
    ...configUtils,
  ***REMOVED***)

  return Object.keys(object).reduce((resolved, key) => {
    resolved[key] = isFunction(object[key]) ? object[key](resolvePath, configUtils) : object[key]

    return resolved
  ***REMOVED***, {***REMOVED***)
***REMOVED***

function extractPluginConfigs(configs) {
  let allConfigs = []

  configs.forEach((config) => {
    allConfigs = [...allConfigs, config]

    const plugins = config?.plugins ?? []

    if (plugins.length === 0) {
      return
    ***REMOVED***

    plugins.forEach((plugin) => {
      if (plugin.__isOptionsFunction) {
        plugin = plugin()
      ***REMOVED***
      allConfigs = [...allConfigs, ...extractPluginConfigs([plugin?.config ?? {***REMOVED***])]
    ***REMOVED***)
  ***REMOVED***)

  return allConfigs
***REMOVED***

function resolveCorePlugins(corePluginConfigs) {
  const result = [...corePluginConfigs].reduceRight((resolved, corePluginConfig) => {
    if (isFunction(corePluginConfig)) {
      return corePluginConfig({ corePlugins: resolved ***REMOVED***)
    ***REMOVED***
    return configurePlugins(corePluginConfig, resolved)
  ***REMOVED***, corePluginList)

  return result
***REMOVED***

function resolvePluginLists(pluginLists) {
  const result = [...pluginLists].reduceRight((resolved, pluginList) => {
    return [...resolved, ...pluginList]
  ***REMOVED***, [])

  return result
***REMOVED***

export default function resolveConfig(configs) {
  let allConfigs = [
    ...extractPluginConfigs(configs),
    {
      prefix: '',
      important: false,
      separator: ':',
    ***REMOVED***,
  ]

  return normalizeConfig(
    defaults(
      {
        theme: resolveFunctionKeys(
          mergeExtensions(mergeThemes(allConfigs.map((t) => t?.theme ?? {***REMOVED***)))
        ),
        corePlugins: resolveCorePlugins(allConfigs.map((c) => c.corePlugins)),
        plugins: resolvePluginLists(configs.map((c) => c?.plugins ?? [])),
      ***REMOVED***,
      ...allConfigs
    )
  )
***REMOVED***
