// renderkid/lib/tools.js
"use strict";

// Generated by CoffeeScript 2.5.1
var cloneDeep, htmlparser, isPlainObject, merge, _objectToDom, self;

htmlparser = require('htmlparser2');

var _require = require('dom-converter');

_objectToDom = _require.objectToDom;
merge = require('lodash/merge');
cloneDeep = require('lodash/cloneDeep');
isPlainObject = require('lodash/isPlainObject');
module.exports = self = {
  repeatString: function repeatString(str, times) {
    var i, j, output, ref;
    output = '';

    for (i = j = 0, ref = times; 0 <= ref ? j < ref : j > ref; i = 0 <= ref ? ++j : --j) {
      output += str;
    ***REMOVED***

    return output;
  ***REMOVED***,
  cloneAndMergeDeep: function cloneAndMergeDeep(base, toAppend) {
    return merge(cloneDeep(base), toAppend);
  ***REMOVED***,
  toDom: function toDom(subject) {
    if (typeof subject === 'string') {
      return self.stringToDom(subject);
    ***REMOVED*** else if (isPlainObject(subject)) {
      return self._objectToDom(subject);
    ***REMOVED*** else {
      throw Error("tools.toDom() only supports strings and objects");
    ***REMOVED***
  ***REMOVED***,
  stringToDom: function stringToDom(string) {
    var handler, parser;
    handler = new htmlparser.DomHandler();
    parser = new htmlparser.Parser(handler);
    parser.write(string);
    parser.end();
    return handler.dom;
  ***REMOVED***,
  _fixQuotesInDom: function _fixQuotesInDom(input) {
    var j, len, node;

    if (Array.isArray(input)) {
      for (j = 0, len = input.length; j < len; j++) {
        node = input[j];

        self._fixQuotesInDom(node);
      ***REMOVED***

      return input;
    ***REMOVED***

    node = input;

    if (node.type === 'text') {
      return node.data = self._quoteNodeText(node.data);
    ***REMOVED*** else {
      return self._fixQuotesInDom(node.children);
    ***REMOVED***
  ***REMOVED***,
  objectToDom: function objectToDom(o) {
    if (!Array.isArray(o)) {
      if (!isPlainObject(o)) {
        throw Error("objectToDom() only accepts a bare object or an array");
      ***REMOVED***
    ***REMOVED***

    return self._fixQuotesInDom(_objectToDom(o));
  ***REMOVED***,
  quote: function quote(str) {
    return String(str).replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/\"/g, '&quot;').replace(/\ /g, '&sp;').replace(/\n/g, '<br />');
  ***REMOVED***,
  _quoteNodeText: function _quoteNodeText(text) {
    return String(text).replace(/\&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/\"/g, '&quot;').replace(/\ /g, '&sp;').replace(/\n/g, "&nl;");
  ***REMOVED***,
  getCols: function getCols() {
    var cols, tty; // Based on https://github.com/jonschlinkert/window-size

    tty = require('tty');

    cols = function () {
      try {
        if (tty.isatty(1) && tty.isatty(2)) {
          if (process.stdout.getWindowSize) {
            return process.stdout.getWindowSize(1)[0];
          ***REMOVED*** else if (tty.getWindowSize) {
            return tty.getWindowSize()[1];
          ***REMOVED*** else if (process.stdout.columns) {
            return process.stdout.columns;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED*** catch (error) {***REMOVED***
    ***REMOVED***();

    if (typeof cols === 'number' && cols > 30) {
      return cols;
    ***REMOVED*** else {
      return 80;
    ***REMOVED***
  ***REMOVED***
***REMOVED***;
