// tailwindcss/scripts/generate-types.js
import prettier from 'prettier'
import { corePlugins ***REMOVED*** from '../src/corePlugins'
import colors from '../src/public/colors'
import defaultTheme from '../src/public/default-theme'
import fs from 'fs'
import path from 'path'
import * as types from './type-utils'

fs.writeFileSync(
  path.join(process.cwd(), 'types', 'generated', 'corePluginList.d.ts'),
  `export type CorePluginList = ${Object.keys(corePlugins)
    .map((p) => `'${p***REMOVED***'`)
    .join(' | ')***REMOVED***`
)

let colorsWithoutDeprecatedColors = Object.fromEntries(
  Object.entries(Object.getOwnPropertyDescriptors(colors))
    .filter(([_, { value ***REMOVED***]) => {
      return typeof value !== 'undefined'
    ***REMOVED***)
    .map(([name, definition]) => [name, definition.value])
)

let deprecatedColors = Object.entries(Object.getOwnPropertyDescriptors(colors))
  .filter(([_, { value ***REMOVED***]) => {
    return typeof value === 'undefined'
  ***REMOVED***)
  .map(([name, definition]) => {
    let warn = console.warn
    let messages = []
    console.warn = (...args) => messages.push(args.pop())
    definition.get()
    console.warn = warn
    let message = messages.join(' ').trim()
    let newColor = message.match(/renamed to `(.*)`/)[1]
    return `/** @deprecated ${message***REMOVED*** */${name***REMOVED***: DefaultColors['${newColor***REMOVED***'],`
  ***REMOVED***)
  .join('\n')

fs.writeFileSync(
  path.join(process.cwd(), 'types', 'generated', 'colors.d.ts'),
  prettier.format(
    `export interface DefaultColors { ${JSON.stringify(colorsWithoutDeprecatedColors).slice(
      1,
      -1
    )***REMOVED***\n${deprecatedColors***REMOVED***\n***REMOVED***`,
    {
      semi: false,
      singleQuote: true,
      printWidth: 100,
      parser: 'typescript',
    ***REMOVED***
  )
)

const defaultThemeTypes = Object.entries(defaultTheme)
  .map(([name, value]) => {
    // Special cases for slightly more accurate types
    if (name === 'keyframes') {
      return [name, `Record<${types.forKeys(value)***REMOVED***, Record<string, CSSDeclarationList>>`]
    ***REMOVED***

    if (name === 'fontSize') {
      return [name, `Record<${types.forKeys(value)***REMOVED***, [string, { lineHeight: string ***REMOVED***]>`]
    ***REMOVED***

    // General cases
    if (typeof value === 'string') {
      return [name, `string`]
    ***REMOVED***

    if (typeof value === 'function') {
      return [name, null]
    ***REMOVED***

    if (typeof value === 'object') {
      if (Object.keys(value).length === 0) {
        return [name, null]
      ***REMOVED***

      return [name, types.forValue(value)]
    ***REMOVED***

    return [name, `unknown`]
  ***REMOVED***)
  .filter(([, type]) => type !== null)
  .map(([name, type]) => `${name***REMOVED***: ${type***REMOVED***`)
  .join('\n')

fs.writeFileSync(
  path.join(process.cwd(), 'types', 'generated', 'default-theme.d.ts'),
  prettier.format(
    `
    type CSSDeclarationList = Record<string, string>
    export type DefaultTheme = { ${defaultThemeTypes***REMOVED*** ***REMOVED***
  `,
    {
      semi: false,
      singleQuote: true,
      printWidth: 100,
      parser: 'typescript',
    ***REMOVED***
  )
)
