// eslint-plugin-flowtype/dist/utilities/spacingFixers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.stripSpacesBefore = exports.stripSpacesAfter = exports.stripSpaces = exports.replaceWithSpaceBefore = exports.replaceWithSpaceAfter = exports.replaceWithSpace = exports.addSpaceBefore = exports.addSpaceAfter = exports.addSpace = void 0;

const stripSpacesBefore = (node, spaces) => {
  return fixer => {
    return fixer.removeRange([node.range[0] - spaces, node.range[0]]);
  ***REMOVED***;
***REMOVED***;

exports.stripSpacesBefore = stripSpacesBefore;

const stripSpacesAfter = (node, spaces) => {
  return fixer => {
    return fixer.removeRange([node.range[1], node.range[1] + spaces]);
  ***REMOVED***;
***REMOVED***;

exports.stripSpacesAfter = stripSpacesAfter;

const addSpaceBefore = node => {
  return fixer => {
    return fixer.insertTextBefore(node, ' ');
  ***REMOVED***;
***REMOVED***;

exports.addSpaceBefore = addSpaceBefore;

const addSpaceAfter = node => {
  return fixer => {
    return fixer.insertTextAfter(node, ' ');
  ***REMOVED***;
***REMOVED***;

exports.addSpaceAfter = addSpaceAfter;

const replaceWithSpaceBefore = (node, spaces) => {
  return fixer => {
    return fixer.replaceTextRange([node.range[0] - spaces, node.range[0]], ' ');
  ***REMOVED***;
***REMOVED***;

exports.replaceWithSpaceBefore = replaceWithSpaceBefore;

const replaceWithSpaceAfter = (node, spaces) => {
  return fixer => {
    return fixer.replaceTextRange([node.range[1], node.range[1] + spaces], ' ');
  ***REMOVED***;
***REMOVED***;

exports.replaceWithSpaceAfter = replaceWithSpaceAfter;

const stripSpaces = (direction, node, spaces) => {
  if (direction === 'before') {
    return stripSpacesBefore(node, spaces);
  ***REMOVED***

  return stripSpacesAfter(node, spaces);
***REMOVED***;

exports.stripSpaces = stripSpaces;

const addSpace = (direction, node) => {
  if (direction === 'before') {
    return addSpaceBefore(node);
  ***REMOVED***

  return addSpaceAfter(node);
***REMOVED***;

exports.addSpace = addSpace;

const replaceWithSpace = (direction, node, spaces) => {
  if (direction === 'before') {
    return replaceWithSpaceBefore(node, spaces);
  ***REMOVED***

  return replaceWithSpaceAfter(node, spaces);
***REMOVED***;

exports.replaceWithSpace = replaceWithSpace;
