// webpack/lib/container/ContainerExposedDependency.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra, Zackary Jackson @ScriptedAlchemy, Marais Rossouw @maraisr
*/

"use strict";

const ModuleDependency = require("../dependencies/ModuleDependency");
const makeSerializable = require("../util/makeSerializable");

/** @typedef {import("../serialization/ObjectMiddleware").ObjectDeserializerContext***REMOVED*** ObjectDeserializerContext */
/** @typedef {import("../serialization/ObjectMiddleware").ObjectSerializerContext***REMOVED*** ObjectSerializerContext */

class ContainerExposedDependency extends ModuleDependency {
	/**
	 * @param {string***REMOVED*** exposedName public name
	 * @param {string***REMOVED*** request request to module
	 */
	constructor(exposedName, request) {
		super(request);
		this.exposedName = exposedName;
	***REMOVED***

	get type() {
		return "container exposed";
	***REMOVED***

	get category() {
		return "esm";
	***REMOVED***

	/**
	 * @returns {string | null***REMOVED*** an identifier to merge equal requests
	 */
	getResourceIdentifier() {
		return `exposed dependency ${this.exposedName***REMOVED***=${this.request***REMOVED***`;
	***REMOVED***

	/**
	 * @param {ObjectSerializerContext***REMOVED*** context context
	 */
	serialize(context) {
		context.write(this.exposedName);
		super.serialize(context);
	***REMOVED***

	/**
	 * @param {ObjectDeserializerContext***REMOVED*** context context
	 */
	deserialize(context) {
		this.exposedName = context.read();
		super.deserialize(context);
	***REMOVED***
***REMOVED***

makeSerializable(
	ContainerExposedDependency,
	"webpack/lib/container/ContainerExposedDependency"
);

module.exports = ContainerExposedDependency;
