// webpack/lib/container/ContainerEntryDependency.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra, Zackary Jackson @ScriptedAlchemy, Marais Rossouw @maraisr
*/

"use strict";

const Dependency = require("../Dependency");
const makeSerializable = require("../util/makeSerializable");

/** @typedef {import("./ContainerEntryModule").ExposeOptions***REMOVED*** ExposeOptions */
/** @typedef {import("./ContainerEntryModule").ExposesList***REMOVED*** ExposesList */

class ContainerEntryDependency extends Dependency {
	/**
	 * @param {string***REMOVED*** name entry name
	 * @param {ExposesList***REMOVED*** exposes list of exposed modules
	 * @param {string***REMOVED*** shareScope name of the share scope
	 */
	constructor(name, exposes, shareScope) {
		super();
		this.name = name;
		this.exposes = exposes;
		this.shareScope = shareScope;
	***REMOVED***

	/**
	 * @returns {string | null***REMOVED*** an identifier to merge equal requests
	 */
	getResourceIdentifier() {
		return `container-entry-${this.name***REMOVED***`;
	***REMOVED***

	get type() {
		return "container entry";
	***REMOVED***

	get category() {
		return "esm";
	***REMOVED***
***REMOVED***

makeSerializable(
	ContainerEntryDependency,
	"webpack/lib/container/ContainerEntryDependency"
);

module.exports = ContainerEntryDependency;
