// @rushstack/eslint-patch/lib/eslint-bulk-suppressions/ast-guards.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.isArrayExpression = isArrayExpression;
exports.isArrowFunctionExpression = isArrowFunctionExpression;
exports.isAssignmentPattern = isAssignmentPattern;
exports.isClassDeclaration = isClassDeclaration;
exports.isClassExpression = isClassExpression;
exports.isExportDefaultDeclaration = isExportDefaultDeclaration;
exports.isExpression = isExpression;
exports.isFunctionDeclaration = isFunctionDeclaration;
exports.isFunctionExpression = isFunctionExpression;
exports.isIdentifier = isIdentifier;
exports.isLiteral = isLiteral;
exports.isMethodDefinition = isMethodDefinition;
exports.isObjectExpression = isObjectExpression;
exports.isPrivateIdentifier = isPrivateIdentifier;
exports.isProperty = isProperty;
exports.isPropertyDefinition = isPropertyDefinition;
exports.isTSEnumDeclaration = isTSEnumDeclaration;
exports.isTSInterfaceDeclaration = isTSInterfaceDeclaration;
exports.isTSModuleDeclaration = isTSModuleDeclaration;
exports.isTSQualifiedName = isTSQualifiedName;
exports.isTSTypeAliasDeclaration = isTSTypeAliasDeclaration;
exports.isVariableDeclarator = isVariableDeclarator;
exports.isClassDeclarationWithName = isClassDeclarationWithName;
exports.isClassPropertyNameNonComputed = isClassPropertyNameNonComputed;
exports.isFunctionDeclarationWithName = isFunctionDeclarationWithName;
exports.isNumberLiteral = isNumberLiteral;
exports.isPropertyNameNonComputed = isPropertyNameNonComputed;
exports.isStringLiteral = isStringLiteral;
exports.isClassExpressionWithName = isClassExpressionWithName;
exports.isFunctionExpressionWithName = isFunctionExpressionWithName;
exports.isNormalAnonymousExpression = isNormalAnonymousExpression;
exports.isNormalAssignmentPattern = isNormalAssignmentPattern;
exports.isNormalClassPropertyDefinition = isNormalClassPropertyDefinition;
exports.isNormalMethodDefinition = isNormalMethodDefinition;
exports.isNormalObjectProperty = isNormalObjectProperty;
exports.isNormalVariableDeclarator = isNormalVariableDeclarator;
exports.isNormalAssignmentPatternWithAnonymousExpressionAssigned = isNormalAssignmentPatternWithAnonymousExpressionAssigned;
exports.isNormalVariableDeclaratorWithAnonymousExpressionAssigned = isNormalVariableDeclaratorWithAnonymousExpressionAssigned;
exports.isNormalObjectPropertyWithAnonymousExpressionAssigned = isNormalObjectPropertyWithAnonymousExpressionAssigned;
exports.isNormalClassPropertyDefinitionWithAnonymousExpressionAssigned = isNormalClassPropertyDefinitionWithAnonymousExpressionAssigned;
exports.isNodeWithName = isNodeWithName;
function isArrayExpression(node) {
    return node.type === 'ArrayExpression';
***REMOVED***
function isArrowFunctionExpression(node) {
    return node.type === 'ArrowFunctionExpression';
***REMOVED***
/** default parameters */
function isAssignmentPattern(node) {
    return node.type === 'AssignmentPattern';
***REMOVED***
function isClassDeclaration(node) {
    return node.type === 'ClassDeclaration';
***REMOVED***
function isClassExpression(node) {
    return node.type === 'ClassExpression';
***REMOVED***
function isExportDefaultDeclaration(node) {
    return node.type === 'ExportDefaultDeclaration';
***REMOVED***
function isExpression(node) {
    return node.type.includes('Expression');
***REMOVED***
function isFunctionDeclaration(node) {
    return node.type === 'FunctionDeclaration';
***REMOVED***
function isFunctionExpression(node) {
    return node.type === 'FunctionExpression';
***REMOVED***
function isIdentifier(node) {
    return node.type === 'Identifier';
***REMOVED***
function isLiteral(node) {
    return node.type === 'Literal';
***REMOVED***
function isMethodDefinition(node) {
    return node.type === 'MethodDefinition';
***REMOVED***
function isObjectExpression(node) {
    return node.type === 'ObjectExpression';
***REMOVED***
function isPrivateIdentifier(node) {
    return node.type === 'PrivateIdentifier';
***REMOVED***
function isProperty(node) {
    return node.type === 'Property';
***REMOVED***
function isPropertyDefinition(node) {
    return node.type === 'PropertyDefinition';
***REMOVED***
function isTSEnumDeclaration(node) {
    return node.type === 'TSEnumDeclaration';
***REMOVED***
function isTSInterfaceDeclaration(node) {
    return node.type === 'TSInterfaceDeclaration';
***REMOVED***
function isTSModuleDeclaration(node) {
    return node.type === 'TSModuleDeclaration';
***REMOVED***
function isTSQualifiedName(node) {
    return node.type === 'TSQualifiedName';
***REMOVED***
function isTSTypeAliasDeclaration(node) {
    return node.type === 'TSTypeAliasDeclaration';
***REMOVED***
function isVariableDeclarator(node) {
    return node.type === 'VariableDeclarator';
***REMOVED***
// Compound Type Guards for @typescript-eslint/types ast-spec compound types
function isClassDeclarationWithName(node) {
    return isClassDeclaration(node) && node.id !== null;
***REMOVED***
function isClassPropertyNameNonComputed(node) {
    return isPrivateIdentifier(node) || isPropertyNameNonComputed(node);
***REMOVED***
function isFunctionDeclarationWithName(node) {
    return isFunctionDeclaration(node) && node.id !== null;
***REMOVED***
function isNumberLiteral(node) {
    return isLiteral(node) && typeof node.value === 'number';
***REMOVED***
function isPropertyNameNonComputed(node) {
    return isIdentifier(node) || isNumberLiteral(node) || isStringLiteral(node);
***REMOVED***
function isStringLiteral(node) {
    return isLiteral(node) && typeof node.value === 'string';
***REMOVED***
function isClassExpressionWithName(node) {
    return isClassExpression(node) && node.id !== null;
***REMOVED***
function isFunctionExpressionWithName(node) {
    return isFunctionExpression(node) && node.id !== null;
***REMOVED***
function isNormalAnonymousExpression(node) {
    const ANONYMOUS_EXPRESSION_GUARDS = [
        isArrowFunctionExpression,
        isClassExpression,
        isFunctionExpression,
        isObjectExpression
    ];
    return ANONYMOUS_EXPRESSION_GUARDS.some((guard) => guard(node));
***REMOVED***
function isNormalAssignmentPattern(node) {
    return isAssignmentPattern(node) && isIdentifier(node.left);
***REMOVED***
function isNormalClassPropertyDefinition(node) {
    return (isPropertyDefinition(node) &&
        (isIdentifier(node.key) || isPrivateIdentifier(node.key)) &&
        node.value !== null);
***REMOVED***
function isNormalMethodDefinition(node) {
    return isMethodDefinition(node) && (isIdentifier(node.key) || isPrivateIdentifier(node.key));
***REMOVED***
function isNormalObjectProperty(node) {
    return isProperty(node) && (isIdentifier(node.key) || isPrivateIdentifier(node.key));
***REMOVED***
function isNormalVariableDeclarator(node) {
    return isVariableDeclarator(node) && isIdentifier(node.id) && node.init !== null;
***REMOVED***
function isNormalAssignmentPatternWithAnonymousExpressionAssigned(node) {
    return isNormalAssignmentPattern(node) && isNormalAnonymousExpression(node.right);
***REMOVED***
function isNormalVariableDeclaratorWithAnonymousExpressionAssigned(node) {
    return isNormalVariableDeclarator(node) && isNormalAnonymousExpression(node.init);
***REMOVED***
function isNormalObjectPropertyWithAnonymousExpressionAssigned(node) {
    return isNormalObjectProperty(node) && isNormalAnonymousExpression(node.value);
***REMOVED***
function isNormalClassPropertyDefinitionWithAnonymousExpressionAssigned(node) {
    return isNormalClassPropertyDefinition(node) && isNormalAnonymousExpression(node.value);
***REMOVED***
function isNodeWithName(node) {
    return (isClassDeclarationWithName(node) ||
        isFunctionDeclarationWithName(node) ||
        isClassExpressionWithName(node) ||
        isFunctionExpressionWithName(node) ||
        isNormalVariableDeclaratorWithAnonymousExpressionAssigned(node) ||
        isNormalObjectPropertyWithAnonymousExpressionAssigned(node) ||
        isNormalClassPropertyDefinitionWithAnonymousExpressionAssigned(node) ||
        isNormalAssignmentPatternWithAnonymousExpressionAssigned(node) ||
        isNormalMethodDefinition(node) ||
        isTSEnumDeclaration(node) ||
        isTSInterfaceDeclaration(node) ||
        isTSTypeAliasDeclaration(node));
***REMOVED***
