// terser/lib/cli.js
import { minify, _default_options ***REMOVED*** from "../main.js";
import { parse ***REMOVED*** from "./parse.js";
import {
    AST_Assign,
    AST_Array,
    AST_Constant,
    AST_Node,
    AST_PropAccess,
    AST_RegExp,
    AST_Sequence,
    AST_Symbol,
    AST_Token,
    walk
***REMOVED*** from "./ast.js";
import { OutputStream ***REMOVED*** from "./output.js";

export async function run_cli({ program, packageJson, fs, path ***REMOVED***) {
    const skip_keys = new Set([ "cname", "parent_scope", "scope", "uses_eval", "uses_with" ]);
    var files = {***REMOVED***;
    var options = {
        compress: false,
        mangle: false
    ***REMOVED***;
    const default_options = await _default_options();
    program.version(packageJson.name + " " + packageJson.version);
    program.parseArgv = program.parse;
    program.parse = undefined;

    if (process.argv.includes("ast")) program.helpInformation = describe_ast;
    else if (process.argv.includes("options")) program.helpInformation = function() {
        var text = [];
        for (var option in default_options) {
            text.push("--" + (option === "sourceMap" ? "source-map" : option) + " options:");
            text.push(format_object(default_options[option]));
            text.push("");
        ***REMOVED***
        return text.join("\n");
    ***REMOVED***;

    program.option("-p, --parse <options>", "Specify parser options.", parse_js());
    program.option("-c, --compress [options]", "Enable compressor/specify compressor options.", parse_js());
    program.option("-m, --mangle [options]", "Mangle names/specify mangler options.", parse_js());
    program.option("--mangle-props [options]", "Mangle properties/specify mangler options.", parse_js());
    program.option("-f, --format [options]", "Format options.", parse_js());
    program.option("-b, --beautify [options]", "Alias for --format.", parse_js());
    program.option("-o, --output <file>", "Output file (default STDOUT).");
    program.option("--comments [filter]", "Preserve copyright comments in the output.");
    program.option("--config-file <file>", "Read minify() options from JSON file.");
    program.option("-d, --define <expr>[=value]", "Global definitions.", parse_js("define"));
    program.option("--ecma <version>", "Specify ECMAScript release: 5, 2015, 2016 or 2017...");
    program.option("-e, --enclose [arg[,...][:value[,...]]]", "Embed output in a big function with configurable arguments and values.");
    program.option("--ie8", "Support non-standard Internet Explorer 8.");
    program.option("--keep-classnames", "Do not mangle/drop class names.");
    program.option("--keep-fnames", "Do not mangle/drop function names. Useful for code relying on Function.prototype.name.");
    program.option("--module", "Input is an ES6 module");
    program.option("--name-cache <file>", "File to hold mangled name mappings.");
    program.option("--rename", "Force symbol expansion.");
    program.option("--no-rename", "Disable symbol expansion.");
    program.option("--safari10", "Support non-standard Safari 10.");
    program.option("--source-map [options]", "Enable source map/specify source map options.", parse_js());
    program.option("--timings", "Display operations run time on STDERR.");
    program.option("--toplevel", "Compress and/or mangle variables in toplevel scope.");
    program.option("--wrap <name>", "Embed everything as a function with “exports” corresponding to “name” globally.");
    program.arguments("[files...]").parseArgv(process.argv);
    if (program.configFile) {
        options = JSON.parse(read_file(program.configFile));
    ***REMOVED***
    if (!program.output && program.sourceMap && program.sourceMap.url != "inline") {
        fatal("ERROR: cannot write source map to STDOUT");
    ***REMOVED***

    [
        "compress",
        "enclose",
        "ie8",
        "mangle",
        "module",
        "safari10",
        "sourceMap",
        "toplevel",
        "wrap"
    ].forEach(function(name) {
        if (name in program) {
            options[name] = program[name];
        ***REMOVED***
    ***REMOVED***);

    if ("ecma" in program) {
        if (program.ecma != (program.ecma | 0)) fatal("ERROR: ecma must be an integer");
        const ecma = program.ecma | 0;
        if (ecma > 5 && ecma < 2015)
            options.ecma = ecma + 2009;
        else
            options.ecma = ecma;
    ***REMOVED***
    if (program.format || program.beautify) {
        const chosenOption = program.format || program.beautify;
        options.format = typeof chosenOption === "object" ? chosenOption : {***REMOVED***;
    ***REMOVED***
    if (program.comments) {
        if (typeof options.format != "object") options.format = {***REMOVED***;
        options.format.comments = typeof program.comments == "string" ? (program.comments == "false" ? false : program.comments) : "some";
    ***REMOVED***
    if (program.define) {
        if (typeof options.compress != "object") options.compress = {***REMOVED***;
        if (typeof options.compress.global_defs != "object") options.compress.global_defs = {***REMOVED***;
        for (var expr in program.define) {
            options.compress.global_defs[expr] = program.define[expr];
        ***REMOVED***
    ***REMOVED***
    if (program.keepClassnames) {
        options.keep_classnames = true;
    ***REMOVED***
    if (program.keepFnames) {
        options.keep_fnames = true;
    ***REMOVED***
    if (program.mangleProps) {
        if (program.mangleProps.domprops) {
            delete program.mangleProps.domprops;
        ***REMOVED*** else {
            if (typeof program.mangleProps != "object") program.mangleProps = {***REMOVED***;
            if (!Array.isArray(program.mangleProps.reserved)) program.mangleProps.reserved = [];
        ***REMOVED***
        if (typeof options.mangle != "object") options.mangle = {***REMOVED***;
        options.mangle.properties = program.mangleProps;
    ***REMOVED***
    if (program.nameCache) {
        options.nameCache = JSON.parse(read_file(program.nameCache, "{***REMOVED***"));
    ***REMOVED***
    if (program.output == "ast") {
        options.format = {
            ast: true,
            code: false
        ***REMOVED***;
    ***REMOVED***
    if (program.parse) {
        if (!program.parse.acorn && !program.parse.spidermonkey) {
            options.parse = program.parse;
        ***REMOVED*** else if (program.sourceMap && program.sourceMap.content == "inline") {
            fatal("ERROR: inline source map only works with built-in parser");
        ***REMOVED***
    ***REMOVED***
    if (~program.rawArgs.indexOf("--rename")) {
        options.rename = true;
    ***REMOVED*** else if (!program.rename) {
        options.rename = false;
    ***REMOVED***

    let convert_path = name => name;
    if (typeof program.sourceMap == "object" && "base" in program.sourceMap) {
        convert_path = function() {
            var base = program.sourceMap.base;
            delete options.sourceMap.base;
            return function(name) {
                return path.relative(base, name);
            ***REMOVED***;
        ***REMOVED***();
    ***REMOVED***

    let filesList;
    if (options.files && options.files.length) {
        filesList = options.files;

        delete options.files;
    ***REMOVED*** else if (program.args.length) {
        filesList = program.args;
    ***REMOVED***

    if (filesList) {
        simple_glob(filesList).forEach(function(name) {
            files[convert_path(name)] = read_file(name);
        ***REMOVED***);
    ***REMOVED*** else {
        await new Promise((resolve) => {
            var chunks = [];
            process.stdin.setEncoding("utf8");
            process.stdin.on("data", function(chunk) {
                chunks.push(chunk);
            ***REMOVED***).on("end", function() {
                files = [ chunks.join("") ];
                resolve();
            ***REMOVED***);
            process.stdin.resume();
        ***REMOVED***);
    ***REMOVED***

    await run_cli();

    function convert_ast(fn) {
        return AST_Node.from_mozilla_ast(Object.keys(files).reduce(fn, null));
    ***REMOVED***

    async function run_cli() {
        var content = program.sourceMap && program.sourceMap.content;
        if (content && content !== "inline") {
            options.sourceMap.content = read_file(content, content);
        ***REMOVED***
        if (program.timings) options.timings = true;

        try {
            if (program.parse) {
                if (program.parse.acorn) {
                    files = convert_ast(function(toplevel, name) {
                        return require("acorn").parse(files[name], {
                            ecmaVersion: 2018,
                            locations: true,
                            program: toplevel,
                            sourceFile: name,
                            sourceType: options.module || program.parse.module ? "module" : "script"
                        ***REMOVED***);
                    ***REMOVED***);
                ***REMOVED*** else if (program.parse.spidermonkey) {
                    files = convert_ast(function(toplevel, name) {
                        var obj = JSON.parse(files[name]);
                        if (!toplevel) return obj;
                        toplevel.body = toplevel.body.concat(obj.body);
                        return toplevel;
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED*** catch (ex) {
            fatal(ex);
        ***REMOVED***

        let result;
        try {
            result = await minify(files, options, fs);
        ***REMOVED*** catch (ex) {
            if (ex.name == "SyntaxError") {
                print_error("Parse error at " + ex.filename + ":" + ex.line + "," + ex.col);
                var col = ex.col;
                var lines = files[ex.filename].split(/\r?\n/);
                var line = lines[ex.line - 1];
                if (!line && !col) {
                    line = lines[ex.line - 2];
                    col = line.length;
                ***REMOVED***
                if (line) {
                    var limit = 70;
                    if (col > limit) {
                        line = line.slice(col - limit);
                        col = limit;
                    ***REMOVED***
                    print_error(line.slice(0, 80));
                    print_error(line.slice(0, col).replace(/\S/g, " ") + "^");
                ***REMOVED***
            ***REMOVED***
            if (ex.defs) {
                print_error("Supported options:");
                print_error(format_object(ex.defs));
            ***REMOVED***
            fatal(ex);
            return;
        ***REMOVED***

        if (program.output == "ast") {
            if (!options.compress && !options.mangle) {
                result.ast.figure_out_scope({***REMOVED***);
            ***REMOVED***
            console.log(JSON.stringify(result.ast, function(key, value) {
                if (value) switch (key) {
                  case "thedef":
                    return symdef(value);
                  case "enclosed":
                    return value.length ? value.map(symdef) : undefined;
                  case "variables":
                  case "globals":
                    return value.size ? collect_from_map(value, symdef) : undefined;
                ***REMOVED***
                if (skip_keys.has(key)) return;
                if (value instanceof AST_Token) return;
                if (value instanceof Map) return;
                if (value instanceof AST_Node) {
                    var result = {
                        _class: "AST_" + value.TYPE
                    ***REMOVED***;
                    if (value.block_scope) {
                        result.variables = value.block_scope.variables;
                        result.enclosed = value.block_scope.enclosed;
                    ***REMOVED***
                    value.CTOR.PROPS.forEach(function(prop) {
                        if (prop !== "block_scope") {
                            result[prop] = value[prop];
                        ***REMOVED***
                    ***REMOVED***);
                    return result;
                ***REMOVED***
                return value;
            ***REMOVED***, 2));
        ***REMOVED*** else if (program.output == "spidermonkey") {
            try {
                const minified = await minify(
                    result.code,
                    {
                        compress: false,
                        mangle: false,
                        format: {
                            ast: true,
                            code: false
                        ***REMOVED***
                    ***REMOVED***,
                    fs
                );
                console.log(JSON.stringify(minified.ast.to_mozilla_ast(), null, 2));
            ***REMOVED*** catch (ex) {
                fatal(ex);
                return;
            ***REMOVED***
        ***REMOVED*** else if (program.output) {
            fs.mkdirSync(path.dirname(program.output), { recursive: true ***REMOVED***);
            fs.writeFileSync(program.output, result.code);
            if (options.sourceMap && options.sourceMap.url !== "inline" && result.map) {
                fs.writeFileSync(program.output + ".map", result.map);
            ***REMOVED***
        ***REMOVED*** else {
            console.log(result.code);
        ***REMOVED***
        if (program.nameCache) {
            fs.writeFileSync(program.nameCache, JSON.stringify(options.nameCache));
        ***REMOVED***
        if (result.timings) for (var phase in result.timings) {
            print_error("- " + phase + ": " + result.timings[phase].toFixed(3) + "s");
        ***REMOVED***
    ***REMOVED***

    function fatal(message) {
        if (message instanceof Error) message = message.stack.replace(/^\S*?Error:/, "ERROR:");
        print_error(message);
        process.exit(1);
    ***REMOVED***

    // A file glob function that only supports "*" and "?" wildcards in the basename.
    // Example: "foo/bar/*baz??.*.js"
    // Argument `glob` may be a string or an array of strings.
    // Returns an array of strings. Garbage in, garbage out.
    function simple_glob(glob) {
        if (Array.isArray(glob)) {
            return [].concat.apply([], glob.map(simple_glob));
        ***REMOVED***
        if (glob && glob.match(/[*?]/)) {
            var dir = path.dirname(glob);
            try {
                var entries = fs.readdirSync(dir);
            ***REMOVED*** catch (ex) {***REMOVED***
            if (entries) {
                var pattern = "^" + path.basename(glob)
                    .replace(/[.+^$[\]\\(){***REMOVED***]/g, "\\$&")
                    .replace(/\*/g, "[^/\\\\]*")
                    .replace(/\?/g, "[^/\\\\]") + "$";
                var mod = process.platform === "win32" ? "i" : "";
                var rx = new RegExp(pattern, mod);
                var results = entries.filter(function(name) {
                    return rx.test(name);
                ***REMOVED***).map(function(name) {
                    return path.join(dir, name);
                ***REMOVED***);
                if (results.length) return results;
            ***REMOVED***
        ***REMOVED***
        return [ glob ];
    ***REMOVED***

    function read_file(path, default_value) {
        try {
            return fs.readFileSync(path, "utf8");
        ***REMOVED*** catch (ex) {
            if ((ex.code == "ENOENT" || ex.code == "ENAMETOOLONG") && default_value != null) return default_value;
            fatal(ex);
        ***REMOVED***
    ***REMOVED***

    function parse_js(flag) {
        return function(value, options) {
            options = options || {***REMOVED***;
            try {
                walk(parse(value, { expression: true ***REMOVED***), node => {
                    if (node instanceof AST_Assign) {
                        var name = node.left.print_to_string();
                        var value = node.right;
                        if (flag) {
                            options[name] = value;
                        ***REMOVED*** else if (value instanceof AST_Array) {
                            options[name] = value.elements.map(to_string);
                        ***REMOVED*** else if (value instanceof AST_RegExp) {
                            value = value.value;
                            options[name] = new RegExp(value.source, value.flags);
                        ***REMOVED*** else {
                            options[name] = to_string(value);
                        ***REMOVED***
                        return true;
                    ***REMOVED***
                    if (node instanceof AST_Symbol || node instanceof AST_PropAccess) {
                        var name = node.print_to_string();
                        options[name] = true;
                        return true;
                    ***REMOVED***
                    if (!(node instanceof AST_Sequence)) throw node;

                    function to_string(value) {
                        return value instanceof AST_Constant ? value.getValue() : value.print_to_string({
                            quote_keys: true
                        ***REMOVED***);
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED*** catch(ex) {
                if (flag) {
                    fatal("Error parsing arguments for '" + flag + "': " + value);
                ***REMOVED*** else {
                    options[value] = null;
                ***REMOVED***
            ***REMOVED***
            return options;
        ***REMOVED***;
    ***REMOVED***

    function symdef(def) {
        var ret = (1e6 + def.id) + " " + def.name;
        if (def.mangled_name) ret += " " + def.mangled_name;
        return ret;
    ***REMOVED***

    function collect_from_map(map, callback) {
        var result = [];
        map.forEach(function (def) {
            result.push(callback(def));
        ***REMOVED***);
        return result;
    ***REMOVED***

    function format_object(obj) {
        var lines = [];
        var padding = "";
        Object.keys(obj).map(function(name) {
            if (padding.length < name.length) padding = Array(name.length + 1).join(" ");
            return [ name, JSON.stringify(obj[name]) ];
        ***REMOVED***).forEach(function(tokens) {
            lines.push("  " + tokens[0] + padding.slice(tokens[0].length - 2) + tokens[1]);
        ***REMOVED***);
        return lines.join("\n");
    ***REMOVED***

    function print_error(msg) {
        process.stderr.write(msg);
        process.stderr.write("\n");
    ***REMOVED***

    function describe_ast() {
        var out = OutputStream({ beautify: true ***REMOVED***);
        function doitem(ctor) {
            out.print("AST_" + ctor.TYPE);
            const props = ctor.SELF_PROPS.filter(prop => !/^\$/.test(prop));

            if (props.length > 0) {
                out.space();
                out.with_parens(function() {
                    props.forEach(function(prop, i) {
                        if (i) out.space();
                        out.print(prop);
                    ***REMOVED***);
                ***REMOVED***);
            ***REMOVED***

            if (ctor.documentation) {
                out.space();
                out.print_string(ctor.documentation);
            ***REMOVED***

            if (ctor.SUBCLASSES.length > 0) {
                out.space();
                out.with_block(function() {
                    ctor.SUBCLASSES.forEach(function(ctor) {
                        out.indent();
                        doitem(ctor);
                        out.newline();
                    ***REMOVED***);
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***
        doitem(AST_Node);
        return out + "\n";
    ***REMOVED***
***REMOVED***
