// workbox-background-sync/lib/QueueDb.js
/*
  Copyright 2021 Google LLC

  Use of this source code is governed by an MIT-style
  license that can be found in the LICENSE file or at
  https://opensource.org/licenses/MIT.
*/
import { openDB ***REMOVED*** from 'idb';
import '../_version.js';
const DB_VERSION = 3;
const DB_NAME = 'workbox-background-sync';
const REQUEST_OBJECT_STORE_NAME = 'requests';
const QUEUE_NAME_INDEX = 'queueName';
/**
 * A class to interact directly an IndexedDB created specifically to save and
 * retrieve QueueStoreEntries. This class encapsulates all the schema details
 * to store the representation of a Queue.
 *
 * @private
 */
export class QueueDb {
    constructor() {
        this._db = null;
    ***REMOVED***
    /**
     * Add QueueStoreEntry to underlying db.
     *
     * @param {UnidentifiedQueueStoreEntry***REMOVED*** entry
     */
    async addEntry(entry) {
        const db = await this.getDb();
        const tx = db.transaction(REQUEST_OBJECT_STORE_NAME, 'readwrite', {
            durability: 'relaxed',
        ***REMOVED***);
        await tx.store.add(entry);
        await tx.done;
    ***REMOVED***
    /**
     * Returns the first entry id in the ObjectStore.
     *
     * @return {number | undefined***REMOVED***
     */
    async getFirstEntryId() {
        const db = await this.getDb();
        const cursor = await db
            .transaction(REQUEST_OBJECT_STORE_NAME)
            .store.openCursor();
        return cursor === null || cursor === void 0 ? void 0 : cursor.value.id;
    ***REMOVED***
    /**
     * Get all the entries filtered by index
     *
     * @param queueName
     * @return {Promise<QueueStoreEntry[]>***REMOVED***
     */
    async getAllEntriesByQueueName(queueName) {
        const db = await this.getDb();
        const results = await db.getAllFromIndex(REQUEST_OBJECT_STORE_NAME, QUEUE_NAME_INDEX, IDBKeyRange.only(queueName));
        return results ? results : new Array();
    ***REMOVED***
    /**
     * Returns the number of entries filtered by index
     *
     * @param queueName
     * @return {Promise<number>***REMOVED***
     */
    async getEntryCountByQueueName(queueName) {
        const db = await this.getDb();
        return db.countFromIndex(REQUEST_OBJECT_STORE_NAME, QUEUE_NAME_INDEX, IDBKeyRange.only(queueName));
    ***REMOVED***
    /**
     * Deletes a single entry by id.
     *
     * @param {number***REMOVED*** id the id of the entry to be deleted
     */
    async deleteEntry(id) {
        const db = await this.getDb();
        await db.delete(REQUEST_OBJECT_STORE_NAME, id);
    ***REMOVED***
    /**
     *
     * @param queueName
     * @returns {Promise<QueueStoreEntry | undefined>***REMOVED***
     */
    async getFirstEntryByQueueName(queueName) {
        return await this.getEndEntryFromIndex(IDBKeyRange.only(queueName), 'next');
    ***REMOVED***
    /**
     *
     * @param queueName
     * @returns {Promise<QueueStoreEntry | undefined>***REMOVED***
     */
    async getLastEntryByQueueName(queueName) {
        return await this.getEndEntryFromIndex(IDBKeyRange.only(queueName), 'prev');
    ***REMOVED***
    /**
     * Returns either the first or the last entries, depending on direction.
     * Filtered by index.
     *
     * @param {IDBCursorDirection***REMOVED*** direction
     * @param {IDBKeyRange***REMOVED*** query
     * @return {Promise<QueueStoreEntry | undefined>***REMOVED***
     * @private
     */
    async getEndEntryFromIndex(query, direction) {
        const db = await this.getDb();
        const cursor = await db
            .transaction(REQUEST_OBJECT_STORE_NAME)
            .store.index(QUEUE_NAME_INDEX)
            .openCursor(query, direction);
        return cursor === null || cursor === void 0 ? void 0 : cursor.value;
    ***REMOVED***
    /**
     * Returns an open connection to the database.
     *
     * @private
     */
    async getDb() {
        if (!this._db) {
            this._db = await openDB(DB_NAME, DB_VERSION, {
                upgrade: this._upgradeDb,
            ***REMOVED***);
        ***REMOVED***
        return this._db;
    ***REMOVED***
    /**
     * Upgrades QueueDB
     *
     * @param {IDBPDatabase<QueueDBSchema>***REMOVED*** db
     * @param {number***REMOVED*** oldVersion
     * @private
     */
    _upgradeDb(db, oldVersion) {
        if (oldVersion > 0 && oldVersion < DB_VERSION) {
            if (db.objectStoreNames.contains(REQUEST_OBJECT_STORE_NAME)) {
                db.deleteObjectStore(REQUEST_OBJECT_STORE_NAME);
            ***REMOVED***
        ***REMOVED***
        const objStore = db.createObjectStore(REQUEST_OBJECT_STORE_NAME, {
            autoIncrement: true,
            keyPath: 'id',
        ***REMOVED***);
        objStore.createIndex(QUEUE_NAME_INDEX, QUEUE_NAME_INDEX, { unique: false ***REMOVED***);
    ***REMOVED***
***REMOVED***
