// html-webpack-plugin/lib/cached-child-compiler.js
// @ts-check
/**
 * @file
 * Helper plugin manages the cached state of the child compilation
 *
 * To optimize performance the child compilation is running asynchronously.
 * Therefore it needs to be started in the compiler.make phase and ends after
 * the compilation.afterCompile phase.
 *
 * To prevent bugs from blocked hooks there is no promise or event based api
 * for this plugin.
 *
 * Example usage:
 *
 * ```js
    const childCompilerPlugin = new PersistentChildCompilerPlugin();
    childCompilerPlugin.addEntry('./src/index.js');
    compiler.hooks.afterCompile.tapAsync('MyPlugin', (compilation, callback) => {
      console.log(childCompilerPlugin.getCompilationResult()['./src/index.js']));
      return true;
    ***REMOVED***);
 * ```
 */
"use strict";

// Import types
/** @typedef {import("webpack").Compiler***REMOVED*** Compiler */
/** @typedef {import("webpack").Compilation***REMOVED*** Compilation */
/** @typedef {import("webpack/lib/FileSystemInfo").Snapshot***REMOVED*** Snapshot */
/** @typedef {import("./child-compiler").ChildCompilationTemplateResult***REMOVED*** ChildCompilationTemplateResult */
/** @typedef {{fileDependencies: string[], contextDependencies: string[], missingDependencies: string[]***REMOVED******REMOVED*** FileDependencies */
/** @typedef {{
  dependencies: FileDependencies,
  compiledEntries: {[entryName: string]: ChildCompilationTemplateResult***REMOVED***
***REMOVED*** | {
  dependencies: FileDependencies,
  error: Error
***REMOVED******REMOVED*** ChildCompilationResult */

const { HtmlWebpackChildCompiler ***REMOVED*** = require("./child-compiler");

/**
 * This plugin is a singleton for performance reasons.
 * To keep track if a plugin does already exist for the compiler they are cached
 * in this map
 * @type {WeakMap<Compiler, PersistentChildCompilerSingletonPlugin>***REMOVED******REMOVED***
 */
const compilerMap = new WeakMap();

class CachedChildCompilation {
  /**
   * @param {Compiler***REMOVED*** compiler
   */
  constructor(compiler) {
    /**
     * @private
     * @type {Compiler***REMOVED***
     */
    this.compiler = compiler;
    // Create a singleton instance for the compiler
    // if there is none
    if (compilerMap.has(compiler)) {
      return;
    ***REMOVED***
    const persistentChildCompilerSingletonPlugin =
      new PersistentChildCompilerSingletonPlugin();
    compilerMap.set(compiler, persistentChildCompilerSingletonPlugin);
    persistentChildCompilerSingletonPlugin.apply(compiler);
  ***REMOVED***

  /**
   * apply is called by the webpack main compiler during the start phase
   * @param {string***REMOVED*** entry
   */
  addEntry(entry) {
    const persistentChildCompilerSingletonPlugin = compilerMap.get(
      this.compiler,
    );
    if (!persistentChildCompilerSingletonPlugin) {
      throw new Error(
        "PersistentChildCompilerSingletonPlugin instance not found.",
      );
    ***REMOVED***
    persistentChildCompilerSingletonPlugin.addEntry(entry);
  ***REMOVED***

  getCompilationResult() {
    const persistentChildCompilerSingletonPlugin = compilerMap.get(
      this.compiler,
    );
    if (!persistentChildCompilerSingletonPlugin) {
      throw new Error(
        "PersistentChildCompilerSingletonPlugin instance not found.",
      );
    ***REMOVED***
    return persistentChildCompilerSingletonPlugin.getLatestResult();
  ***REMOVED***

  /**
   * Returns the result for the given entry
   * @param {string***REMOVED*** entry
   * @returns {
      | { mainCompilationHash: string, error: Error ***REMOVED***
      | { mainCompilationHash: string, compiledEntry: ChildCompilationTemplateResult ***REMOVED***
    ***REMOVED***
   */
  getCompilationEntryResult(entry) {
    const latestResult = this.getCompilationResult();
    const compilationResult = latestResult.compilationResult;
    return "error" in compilationResult
      ? {
          mainCompilationHash: latestResult.mainCompilationHash,
          error: compilationResult.error,
        ***REMOVED***
      : {
          mainCompilationHash: latestResult.mainCompilationHash,
          compiledEntry: compilationResult.compiledEntries[entry],
        ***REMOVED***;
  ***REMOVED***
***REMOVED***

class PersistentChildCompilerSingletonPlugin {
  /**
   *
   * @param {{fileDependencies: string[], contextDependencies: string[], missingDependencies: string[]***REMOVED******REMOVED*** fileDependencies
   * @param {Compilation***REMOVED*** mainCompilation
   * @param {number***REMOVED*** startTime
   */
  static createSnapshot(fileDependencies, mainCompilation, startTime) {
    return new Promise((resolve, reject) => {
      mainCompilation.fileSystemInfo.createSnapshot(
        startTime,
        fileDependencies.fileDependencies,
        fileDependencies.contextDependencies,
        fileDependencies.missingDependencies,
        // @ts-ignore
        null,
        (err, snapshot) => {
          if (err) {
            return reject(err);
          ***REMOVED***
          resolve(snapshot);
        ***REMOVED***,
      );
    ***REMOVED***);
  ***REMOVED***

  /**
   * Returns true if the files inside this snapshot
   * have not been changed
   *
   * @param {Snapshot***REMOVED*** snapshot
   * @param {Compilation***REMOVED*** mainCompilation
   * @returns {Promise<boolean | undefined>***REMOVED***
   */
  static isSnapshotValid(snapshot, mainCompilation) {
    return new Promise((resolve, reject) => {
      mainCompilation.fileSystemInfo.checkSnapshotValid(
        snapshot,
        (err, isValid) => {
          if (err) {
            reject(err);
          ***REMOVED***
          resolve(isValid);
        ***REMOVED***,
      );
    ***REMOVED***);
  ***REMOVED***

  static watchFiles(mainCompilation, fileDependencies) {
    Object.keys(fileDependencies).forEach((dependencyType) => {
      fileDependencies[dependencyType].forEach((fileDependency) => {
        mainCompilation[dependencyType].add(fileDependency);
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***

  constructor() {
    /**
     * @private
     * @type {
      | {
        isCompiling: false,
        isVerifyingCache: false,
        entries: string[],
        compiledEntries: string[],
        mainCompilationHash: string,
        compilationResult: ChildCompilationResult
      ***REMOVED***
    | Readonly<{
      isCompiling: false,
      isVerifyingCache: true,
      entries: string[],
      previousEntries: string[],
      previousResult: ChildCompilationResult
    ***REMOVED***>
    | Readonly <{
      isVerifyingCache: false,
      isCompiling: true,
      entries: string[],
    ***REMOVED***>
  ***REMOVED*** the internal compilation state */
    this.compilationState = {
      isCompiling: false,
      isVerifyingCache: false,
      entries: [],
      compiledEntries: [],
      mainCompilationHash: "initial",
      compilationResult: {
        dependencies: {
          fileDependencies: [],
          contextDependencies: [],
          missingDependencies: [],
        ***REMOVED***,
        compiledEntries: {***REMOVED***,
      ***REMOVED***,
    ***REMOVED***;
  ***REMOVED***

  /**
   * apply is called by the webpack main compiler during the start phase
   * @param {Compiler***REMOVED*** compiler
   */
  apply(compiler) {
    /** @type Promise<ChildCompilationResult> */
    let childCompilationResultPromise = Promise.resolve({
      dependencies: {
        fileDependencies: [],
        contextDependencies: [],
        missingDependencies: [],
      ***REMOVED***,
      compiledEntries: {***REMOVED***,
    ***REMOVED***);
    /**
     * The main compilation hash which will only be updated
     * if the childCompiler changes
     */
    /** @type {string***REMOVED*** */
    let mainCompilationHashOfLastChildRecompile = "";
    /** @type {Snapshot | undefined***REMOVED*** */
    let previousFileSystemSnapshot;
    let compilationStartTime = new Date().getTime();

    compiler.hooks.make.tapAsync(
      "PersistentChildCompilerSingletonPlugin",
      (mainCompilation, callback) => {
        if (
          this.compilationState.isCompiling ||
          this.compilationState.isVerifyingCache
        ) {
          return callback(new Error("Child compilation has already started"));
        ***REMOVED***

        // Update the time to the current compile start time
        compilationStartTime = new Date().getTime();

        // The compilation starts - adding new templates is now not possible anymore
        this.compilationState = {
          isCompiling: false,
          isVerifyingCache: true,
          previousEntries: this.compilationState.compiledEntries,
          previousResult: this.compilationState.compilationResult,
          entries: this.compilationState.entries,
        ***REMOVED***;

        // Validate cache:
        const isCacheValidPromise = this.isCacheValid(
          previousFileSystemSnapshot,
          mainCompilation,
        );

        let cachedResult = childCompilationResultPromise;
        childCompilationResultPromise = isCacheValidPromise.then(
          (isCacheValid) => {
            // Reuse cache
            if (isCacheValid) {
              return cachedResult;
            ***REMOVED***
            // Start the compilation
            const compiledEntriesPromise = this.compileEntries(
              mainCompilation,
              this.compilationState.entries,
            );
            // Update snapshot as soon as we know the fileDependencies
            // this might possibly cause bugs if files were changed between
            // compilation start and snapshot creation
            compiledEntriesPromise
              .then((childCompilationResult) => {
                return PersistentChildCompilerSingletonPlugin.createSnapshot(
                  childCompilationResult.dependencies,
                  mainCompilation,
                  compilationStartTime,
                );
              ***REMOVED***)
              .then((snapshot) => {
                previousFileSystemSnapshot = snapshot;
              ***REMOVED***);
            return compiledEntriesPromise;
          ***REMOVED***,
        );

        // Add files to compilation which needs to be watched:
        mainCompilation.hooks.optimizeTree.tapAsync(
          "PersistentChildCompilerSingletonPlugin",
          (chunks, modules, callback) => {
            const handleCompilationDonePromise =
              childCompilationResultPromise.then((childCompilationResult) => {
                this.watchFiles(
                  mainCompilation,
                  childCompilationResult.dependencies,
                );
              ***REMOVED***);
            handleCompilationDonePromise.then(
              // @ts-ignore
              () => callback(null, chunks, modules),
              callback,
            );
          ***REMOVED***,
        );

        // Store the final compilation once the main compilation hash is known
        mainCompilation.hooks.additionalAssets.tapAsync(
          "PersistentChildCompilerSingletonPlugin",
          (callback) => {
            const didRecompilePromise = Promise.all([
              childCompilationResultPromise,
              cachedResult,
            ]).then(([childCompilationResult, cachedResult]) => {
              // Update if childCompilation changed
              return cachedResult !== childCompilationResult;
            ***REMOVED***);

            const handleCompilationDonePromise = Promise.all([
              childCompilationResultPromise,
              didRecompilePromise,
            ]).then(([childCompilationResult, didRecompile]) => {
              // Update hash and snapshot if childCompilation changed
              if (didRecompile) {
                mainCompilationHashOfLastChildRecompile =
                  /** @type {string***REMOVED*** */ (mainCompilation.hash);
              ***REMOVED***
              this.compilationState = {
                isCompiling: false,
                isVerifyingCache: false,
                entries: this.compilationState.entries,
                compiledEntries: this.compilationState.entries,
                compilationResult: childCompilationResult,
                mainCompilationHash: mainCompilationHashOfLastChildRecompile,
              ***REMOVED***;
            ***REMOVED***);
            handleCompilationDonePromise.then(() => callback(null), callback);
          ***REMOVED***,
        );

        // Continue compilation:
        callback(null);
      ***REMOVED***,
    );
  ***REMOVED***

  /**
   * Add a new entry to the next compile run
   * @param {string***REMOVED*** entry
   */
  addEntry(entry) {
    if (
      this.compilationState.isCompiling ||
      this.compilationState.isVerifyingCache
    ) {
      throw new Error(
        "The child compiler has already started to compile. " +
          "Please add entries before the main compiler 'make' phase has started or " +
          "after the compilation is done.",
      );
    ***REMOVED***
    if (this.compilationState.entries.indexOf(entry) === -1) {
      this.compilationState.entries = [...this.compilationState.entries, entry];
    ***REMOVED***
  ***REMOVED***

  getLatestResult() {
    if (
      this.compilationState.isCompiling ||
      this.compilationState.isVerifyingCache
    ) {
      throw new Error(
        "The child compiler is not done compiling. " +
          "Please access the result after the compiler 'make' phase has started or " +
          "after the compilation is done.",
      );
    ***REMOVED***
    return {
      mainCompilationHash: this.compilationState.mainCompilationHash,
      compilationResult: this.compilationState.compilationResult,
    ***REMOVED***;
  ***REMOVED***

  /**
   * Verify that the cache is still valid
   * @private
   * @param {Snapshot | undefined***REMOVED*** snapshot
   * @param {Compilation***REMOVED*** mainCompilation
   * @returns {Promise<boolean | undefined>***REMOVED***
   */
  isCacheValid(snapshot, mainCompilation) {
    if (!this.compilationState.isVerifyingCache) {
      return Promise.reject(
        new Error(
          "Cache validation can only be done right before the compilation starts",
        ),
      );
    ***REMOVED***
    // If there are no entries we don't need a new child compilation
    if (this.compilationState.entries.length === 0) {
      return Promise.resolve(true);
    ***REMOVED***
    // If there are new entries the cache is invalid
    if (
      this.compilationState.entries !== this.compilationState.previousEntries
    ) {
      return Promise.resolve(false);
    ***REMOVED***
    // Mark the cache as invalid if there is no snapshot
    if (!snapshot) {
      return Promise.resolve(false);
    ***REMOVED***

    return PersistentChildCompilerSingletonPlugin.isSnapshotValid(
      snapshot,
      mainCompilation,
    );
  ***REMOVED***

  /**
   * Start to compile all templates
   *
   * @private
   * @param {Compilation***REMOVED*** mainCompilation
   * @param {string[]***REMOVED*** entries
   * @returns {Promise<ChildCompilationResult>***REMOVED***
   */
  compileEntries(mainCompilation, entries) {
    const compiler = new HtmlWebpackChildCompiler(entries);
    return compiler.compileTemplates(mainCompilation).then(
      (result) => {
        return {
          // The compiled sources to render the content
          compiledEntries: result,
          // The file dependencies to find out if a
          // recompilation is required
          dependencies: compiler.fileDependencies,
          // The main compilation hash can be used to find out
          // if this compilation was done during the current compilation
          mainCompilationHash: mainCompilation.hash,
        ***REMOVED***;
      ***REMOVED***,
      (error) => ({
        // The compiled sources to render the content
        error,
        // The file dependencies to find out if a
        // recompilation is required
        dependencies: compiler.fileDependencies,
        // The main compilation hash can be used to find out
        // if this compilation was done during the current compilation
        mainCompilationHash: mainCompilation.hash,
      ***REMOVED***),
    );
  ***REMOVED***

  /**
   * @private
   * @param {Compilation***REMOVED*** mainCompilation
   * @param {FileDependencies***REMOVED*** files
   */
  watchFiles(mainCompilation, files) {
    PersistentChildCompilerSingletonPlugin.watchFiles(mainCompilation, files);
  ***REMOVED***
***REMOVED***

module.exports = {
  CachedChildCompilation,
***REMOVED***;
