// html-webpack-plugin/lib/html-tags.js
// @ts-check
/** @typedef {import("../typings").HtmlTagObject***REMOVED*** HtmlTagObject */
/**
 * @file
 * This file provides to helper to create html as a object representation as
 * those objects are easier to modify than pure string representations
 *
 * Usage:
 * ```
 * const element = createHtmlTagObject('h1', {class: 'demo'***REMOVED***, 'Hello World');
 * const html = htmlTagObjectToString(element);
 * console.log(html) // -> <h1 class="demo">Hello World</h1>
 * ```
 */

/**
 * All html tag elements which must not contain innerHTML
 * @see https://www.w3.org/TR/html5/syntax.html#void-elements
 */
const voidTags = [
  "area",
  "base",
  "br",
  "col",
  "embed",
  "hr",
  "img",
  "input",
  "keygen",
  "link",
  "meta",
  "param",
  "source",
  "track",
  "wbr",
];

/**
 * Turn a tag definition into a html string
 * @param {HtmlTagObject***REMOVED*** tagDefinition
 *  A tag element according to the htmlWebpackPlugin object notation
 *
 * @param xhtml {boolean***REMOVED***
 *   Whether the generated html should add closing slashes to be xhtml compliant
 */
function htmlTagObjectToString(tagDefinition, xhtml) {
  const attributes = Object.keys(tagDefinition.attributes || {***REMOVED***)
    .filter(function (attributeName) {
      return (
        tagDefinition.attributes[attributeName] === "" ||
        tagDefinition.attributes[attributeName]
      );
    ***REMOVED***)
    .map(function (attributeName) {
      if (tagDefinition.attributes[attributeName] === true) {
        return xhtml
          ? attributeName + '="' + attributeName + '"'
          : attributeName;
      ***REMOVED***
      return (
        attributeName + '="' + tagDefinition.attributes[attributeName] + '"'
      );
    ***REMOVED***);
  return (
    "<" +
    [tagDefinition.tagName].concat(attributes).join(" ") +
    (tagDefinition.voidTag && xhtml ? "/" : "") +
    ">" +
    (tagDefinition.innerHTML || "") +
    (tagDefinition.voidTag ? "" : "</" + tagDefinition.tagName + ">")
  );
***REMOVED***

/**
 * Static helper to create a tag object to be get injected into the dom
 *
 * @param {string***REMOVED*** tagName
 * the name of the tag e.g. 'div'
 *
 * @param {{[attributeName: string]: string|boolean|null|undefined***REMOVED******REMOVED*** [attributes]
 * tag attributes e.g. `{ 'class': 'example', disabled: true ***REMOVED***`
 *
 * @param {string***REMOVED*** [innerHTML]
 *
 * @param {{[attributeName: string]: string|boolean|null|undefined***REMOVED******REMOVED*** [meta]
 * meta information about the tag e.g. `{ 'plugin': 'html-webpack-plugin' ***REMOVED***`
 *
 * @returns {HtmlTagObject***REMOVED***
 */
function createHtmlTagObject(tagName, attributes, innerHTML, meta) {
  return {
    tagName: tagName,
    voidTag: voidTags.indexOf(tagName) !== -1,
    attributes: attributes || {***REMOVED***,
    meta: meta || {***REMOVED***,
    innerHTML: innerHTML,
  ***REMOVED***;
***REMOVED***

/**
 * The `HtmlTagArray Array with a custom `.toString()` method.
 *
 * This allows the following:
 * ```
 *   const tags = HtmlTagArray.from([tag1, tag2]);
 *   const scriptTags = tags.filter((tag) => tag.tagName === 'script');
 *   const html = scriptTags.toString();
 * ```
 *
 * Or inside a string literal:
 * ```
 *   const tags = HtmlTagArray.from([tag1, tag2]);
 *   const html = `<html><body>${tags.filter((tag) => tag.tagName === 'script')***REMOVED***</body></html>`;
 * ```
 *
 */
class HtmlTagArray extends Array {
  toString() {
    return this.join("");
  ***REMOVED***
***REMOVED***

module.exports = {
  HtmlTagArray: HtmlTagArray,
  createHtmlTagObject: createHtmlTagObject,
  htmlTagObjectToString: htmlTagObjectToString,
***REMOVED***;
