// fork-ts-checker-webpack-plugin/lib/rpc/RpcMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
function createRpcMessage(procedure, id, type, payload, source) {
    return {
        rpc: true,
        type,
        id,
        procedure,
        payload,
        source,
    ***REMOVED***;
***REMOVED***
exports.createRpcMessage = createRpcMessage;
function createRpcCall(procedure, index, payload) {
    return createRpcMessage(procedure, index, 'call', payload);
***REMOVED***
exports.createRpcCall = createRpcCall;
function createRpcReturn(procedure, index, payload) {
    return createRpcMessage(procedure, index, 'return', payload);
***REMOVED***
exports.createRpcReturn = createRpcReturn;
function createRpcThrow(procedure, index, payload) {
    return createRpcMessage(procedure, index, 'throw', payload);
***REMOVED***
exports.createRpcThrow = createRpcThrow;
function isRpcMessage(candidate) {
    return !!(typeof candidate === 'object' && candidate && candidate.rpc);
***REMOVED***
exports.isRpcMessage = isRpcMessage;
function isRpcCallMessage(candidate) {
    return isRpcMessage(candidate) && candidate.type === 'call';
***REMOVED***
exports.isRpcCallMessage = isRpcCallMessage;
function isRpcReturnMessage(candidate) {
    return isRpcMessage(candidate) && candidate.type === 'return';
***REMOVED***
exports.isRpcReturnMessage = isRpcReturnMessage;
function isRpcThrowMessage(candidate) {
    return isRpcMessage(candidate) && candidate.type === 'throw';
***REMOVED***
exports.isRpcThrowMessage = isRpcThrowMessage;
function getRpcMessageKey(message) {
    return `${message.procedure***REMOVED***_${message.id***REMOVED***`;
***REMOVED***
exports.getRpcMessageKey = getRpcMessageKey;
