// css-tree/lib/lexer/prepare-tokens.js
var tokenize = require('../tokenizer');
var TokenStream = require('../common/TokenStream');
var tokenStream = new TokenStream();
var astToTokens = {
    decorator: function(handlers) {
        var curNode = null;
        var prev = { len: 0, node: null ***REMOVED***;
        var nodes = [prev];
        var buffer = '';

        return {
            children: handlers.children,
            node: function(node) {
                var tmp = curNode;
                curNode = node;
                handlers.node.call(this, node);
                curNode = tmp;
            ***REMOVED***,
            chunk: function(chunk) {
                buffer += chunk;
                if (prev.node !== curNode) {
                    nodes.push({
                        len: chunk.length,
                        node: curNode
                    ***REMOVED***);
                ***REMOVED*** else {
                    prev.len += chunk.length;
                ***REMOVED***
            ***REMOVED***,
            result: function() {
                return prepareTokens(buffer, nodes);
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;

function prepareTokens(str, nodes) {
    var tokens = [];
    var nodesOffset = 0;
    var nodesIndex = 0;
    var currentNode = nodes ? nodes[nodesIndex].node : null;

    tokenize(str, tokenStream);

    while (!tokenStream.eof) {
        if (nodes) {
            while (nodesIndex < nodes.length && nodesOffset + nodes[nodesIndex].len <= tokenStream.tokenStart) {
                nodesOffset += nodes[nodesIndex++].len;
                currentNode = nodes[nodesIndex].node;
            ***REMOVED***
        ***REMOVED***

        tokens.push({
            type: tokenStream.tokenType,
            value: tokenStream.getTokenValue(),
            index: tokenStream.tokenIndex, // TODO: remove it, temporary solution
            balance: tokenStream.balance[tokenStream.tokenIndex], // TODO: remove it, temporary solution
            node: currentNode
        ***REMOVED***);
        tokenStream.next();
        // console.log({ ...tokens[tokens.length - 1], node: undefined ***REMOVED***);
    ***REMOVED***

    return tokens;
***REMOVED***

module.exports = function(value, syntax) {
    if (typeof value === 'string') {
        return prepareTokens(value, null);
    ***REMOVED***

    return syntax.generate(value, astToTokens);
***REMOVED***;
