// ajv-keywords/dist/definitions/deepRequired.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const codegen_1 = require("ajv/dist/compile/codegen");
function getDef() {
    return {
        keyword: "deepRequired",
        type: "object",
        schemaType: "array",
        code(ctx) {
            const { schema, data ***REMOVED*** = ctx;
            const props = schema.map((jp) => (0, codegen_1._) `(${getData(jp)***REMOVED***) === undefined`);
            ctx.fail((0, codegen_1.or)(...props));
            function getData(jsonPointer) {
                if (jsonPointer === "")
                    throw new Error("empty JSON pointer not allowed");
                const segments = jsonPointer.split("/");
                let x = data;
                const xs = segments.map((s, i) => i ? (x = (0, codegen_1._) `${x***REMOVED***${(0, codegen_1.getProperty)(unescapeJPSegment(s))***REMOVED***`) : x);
                return (0, codegen_1.and)(...xs);
            ***REMOVED***
        ***REMOVED***,
        metaSchema: {
            type: "array",
            items: { type: "string", format: "json-pointer" ***REMOVED***,
        ***REMOVED***,
    ***REMOVED***;
***REMOVED***
exports.default = getDef;
function unescapeJPSegment(s) {
    return s.replace(/~1/g, "/").replace(/~0/g, "~");
***REMOVED***
module.exports = getDef;
