// ajv-keywords/dist/definitions/deepProperties.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const _util_1 = require("./_util");
function getDef(opts) {
    return {
        keyword: "deepProperties",
        type: "object",
        schemaType: "object",
        macro: function (schema) {
            const allOf = [];
            for (const pointer in schema)
                allOf.push(getSchema(pointer, schema[pointer]));
            return { allOf ***REMOVED***;
        ***REMOVED***,
        metaSchema: {
            type: "object",
            propertyNames: { type: "string", format: "json-pointer" ***REMOVED***,
            additionalProperties: (0, _util_1.metaSchemaRef)(opts),
        ***REMOVED***,
    ***REMOVED***;
***REMOVED***
exports.default = getDef;
function getSchema(jsonPointer, schema) {
    const segments = jsonPointer.split("/");
    const rootSchema = {***REMOVED***;
    let pointerSchema = rootSchema;
    for (let i = 1; i < segments.length; i++) {
        let segment = segments[i];
        const isLast = i === segments.length - 1;
        segment = unescapeJsonPointer(segment);
        const properties = (pointerSchema.properties = {***REMOVED***);
        let items;
        if (/[0-9]+/.test(segment)) {
            let count = +segment;
            items = pointerSchema.items = [];
            pointerSchema.type = ["object", "array"];
            while (count--)
                items.push({***REMOVED***);
        ***REMOVED***
        else {
            pointerSchema.type = "object";
        ***REMOVED***
        pointerSchema = isLast ? schema : {***REMOVED***;
        properties[segment] = pointerSchema;
        if (items)
            items.push(pointerSchema);
    ***REMOVED***
    return rootSchema;
***REMOVED***
function unescapeJsonPointer(str) {
    return str.replace(/~1/g, "/").replace(/~0/g, "~");
***REMOVED***
module.exports = getDef;
