// ajv-keywords/dist/definitions/instanceof.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const CONSTRUCTORS = {
    Object,
    Array,
    Function,
    Number,
    String,
    Date,
    RegExp,
***REMOVED***;
/* istanbul ignore else */
if (typeof Buffer != "undefined")
    CONSTRUCTORS.Buffer = Buffer;
/* istanbul ignore else */
if (typeof Promise != "undefined")
    CONSTRUCTORS.Promise = Promise;
const getDef = Object.assign(_getDef, { CONSTRUCTORS ***REMOVED***);
function _getDef() {
    return {
        keyword: "instanceof",
        schemaType: ["string", "array"],
        compile(schema) {
            if (typeof schema == "string") {
                const C = getConstructor(schema);
                return (data) => data instanceof C;
            ***REMOVED***
            if (Array.isArray(schema)) {
                const constructors = schema.map(getConstructor);
                return (data) => {
                    for (const C of constructors) {
                        if (data instanceof C)
                            return true;
                    ***REMOVED***
                    return false;
                ***REMOVED***;
            ***REMOVED***
            /* istanbul ignore next */
            throw new Error("ajv implementation error");
        ***REMOVED***,
        metaSchema: {
            anyOf: [{ type: "string" ***REMOVED***, { type: "array", items: { type: "string" ***REMOVED*** ***REMOVED***],
        ***REMOVED***,
    ***REMOVED***;
***REMOVED***
function getConstructor(c) {
    const C = CONSTRUCTORS[c];
    if (C)
        return C;
    throw new Error(`invalid "instanceof" keyword value ${c***REMOVED***`);
***REMOVED***
exports.default = getDef;
module.exports = getDef;
