// ajv-keywords/dist/definitions/patternRequired.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const codegen_1 = require("ajv/dist/compile/codegen");
const _util_1 = require("./_util");
const error = {
    message: ({ params: { missingPattern ***REMOVED*** ***REMOVED***) => (0, codegen_1.str) `should have property matching pattern '${missingPattern***REMOVED***'`,
    params: ({ params: { missingPattern ***REMOVED*** ***REMOVED***) => (0, codegen_1._) `{missingPattern: ${missingPattern***REMOVED******REMOVED***`,
***REMOVED***;
function getDef() {
    return {
        keyword: "patternRequired",
        type: "object",
        schemaType: "array",
        error,
        code(cxt) {
            const { gen, schema, data ***REMOVED*** = cxt;
            if (schema.length === 0)
                return;
            const valid = gen.let("valid", true);
            for (const pat of schema)
                validateProperties(pat);
            function validateProperties(pattern) {
                const matched = gen.let("matched", false);
                gen.forIn("key", data, (key) => {
                    gen.assign(matched, (0, codegen_1._) `${(0, _util_1.usePattern)(cxt, pattern)***REMOVED***.test(${key***REMOVED***)`);
                    gen.if(matched, () => gen.break());
                ***REMOVED***);
                cxt.setParams({ missingPattern: pattern ***REMOVED***);
                gen.assign(valid, (0, codegen_1.and)(valid, matched));
                cxt.pass(valid);
            ***REMOVED***
        ***REMOVED***,
        metaSchema: {
            type: "array",
            items: { type: "string", format: "regex" ***REMOVED***,
            uniqueItems: true,
        ***REMOVED***,
    ***REMOVED***;
***REMOVED***
exports.default = getDef;
module.exports = getDef;
