// ajv-keywords/dist/definitions/regexp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const codegen_1 = require("ajv/dist/compile/codegen");
const _util_1 = require("./_util");
const regexpMetaSchema = {
    type: "object",
    properties: {
        pattern: { type: "string" ***REMOVED***,
        flags: { type: "string", nullable: true ***REMOVED***,
    ***REMOVED***,
    required: ["pattern"],
    additionalProperties: false,
***REMOVED***;
const metaRegexp = /^\/(.*)\/([gimuy]*)$/;
function getDef() {
    return {
        keyword: "regexp",
        type: "string",
        schemaType: ["string", "object"],
        code(cxt) {
            const { data, schema ***REMOVED*** = cxt;
            const regx = getRegExp(schema);
            cxt.pass((0, codegen_1._) `${regx***REMOVED***.test(${data***REMOVED***)`);
            function getRegExp(sch) {
                if (typeof sch == "object")
                    return (0, _util_1.usePattern)(cxt, sch.pattern, sch.flags);
                const rx = metaRegexp.exec(sch);
                if (rx)
                    return (0, _util_1.usePattern)(cxt, rx[1], rx[2]);
                throw new Error("cannot parse string into RegExp");
            ***REMOVED***
        ***REMOVED***,
        metaSchema: {
            anyOf: [{ type: "string" ***REMOVED***, regexpMetaSchema],
        ***REMOVED***,
    ***REMOVED***;
***REMOVED***
exports.default = getDef;
module.exports = getDef;
