// ajv-keywords/dist/definitions/select.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const codegen_1 = require("ajv/dist/compile/codegen");
const _util_1 = require("./_util");
const error = {
    message: ({ params: { schemaProp ***REMOVED*** ***REMOVED***) => schemaProp
        ? (0, codegen_1.str) `should match case "${schemaProp***REMOVED***" schema`
        : (0, codegen_1.str) `should match default case schema`,
    params: ({ params: { schemaProp ***REMOVED*** ***REMOVED***) => schemaProp ? (0, codegen_1._) `{failingCase: ${schemaProp***REMOVED******REMOVED***` : (0, codegen_1._) `{failingDefault: true***REMOVED***`,
***REMOVED***;
function getDef(opts) {
    const metaSchema = (0, _util_1.metaSchemaRef)(opts);
    return [
        {
            keyword: "select",
            schemaType: ["string", "number", "boolean", "null"],
            $data: true,
            error,
            dependencies: ["selectCases"],
            code(cxt) {
                const { gen, schemaCode, parentSchema ***REMOVED*** = cxt;
                cxt.block$data(codegen_1.nil, () => {
                    const valid = gen.let("valid", true);
                    const schValid = gen.name("_valid");
                    const value = gen.const("value", (0, codegen_1._) `${schemaCode***REMOVED*** === null ? "null" : ${schemaCode***REMOVED***`);
                    gen.if(false); // optimizer should remove it from generated code
                    for (const schemaProp in parentSchema.selectCases) {
                        cxt.setParams({ schemaProp ***REMOVED***);
                        gen.elseIf((0, codegen_1._) `"" + ${value***REMOVED*** == ${schemaProp***REMOVED***`); // intentional ==, to match numbers and booleans
                        const schCxt = cxt.subschema({ keyword: "selectCases", schemaProp ***REMOVED***, schValid);
                        cxt.mergeEvaluated(schCxt, codegen_1.Name);
                        gen.assign(valid, schValid);
                    ***REMOVED***
                    gen.else();
                    if (parentSchema.selectDefault !== undefined) {
                        cxt.setParams({ schemaProp: undefined ***REMOVED***);
                        const schCxt = cxt.subschema({ keyword: "selectDefault" ***REMOVED***, schValid);
                        cxt.mergeEvaluated(schCxt, codegen_1.Name);
                        gen.assign(valid, schValid);
                    ***REMOVED***
                    gen.endIf();
                    cxt.pass(valid);
                ***REMOVED***);
            ***REMOVED***,
        ***REMOVED***,
        {
            keyword: "selectCases",
            dependencies: ["select"],
            metaSchema: {
                type: "object",
                additionalProperties: metaSchema,
            ***REMOVED***,
        ***REMOVED***,
        {
            keyword: "selectDefault",
            dependencies: ["select", "selectCases"],
            metaSchema,
        ***REMOVED***,
    ];
***REMOVED***
exports.default = getDef;
module.exports = getDef;
