// ajv-keywords/dist/definitions/typeof.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const codegen_1 = require("ajv/dist/compile/codegen");
const TYPES = ["undefined", "string", "number", "object", "function", "boolean", "symbol"];
function getDef() {
    return {
        keyword: "typeof",
        schemaType: ["string", "array"],
        code(cxt) {
            const { data, schema, schemaValue ***REMOVED*** = cxt;
            cxt.fail(typeof schema == "string"
                ? (0, codegen_1._) `typeof ${data***REMOVED*** != ${schema***REMOVED***`
                : (0, codegen_1._) `${schemaValue***REMOVED***.indexOf(typeof ${data***REMOVED***) < 0`);
        ***REMOVED***,
        metaSchema: {
            anyOf: [
                { type: "string", enum: TYPES ***REMOVED***,
                { type: "array", items: { type: "string", enum: TYPES ***REMOVED*** ***REMOVED***,
            ],
        ***REMOVED***,
    ***REMOVED***;
***REMOVED***
exports.default = getDef;
module.exports = getDef;
