// @babel/plugin-transform-runtime/src/get-runtime-path/index.ts
import path from "node:path";

import { createRequire ***REMOVED*** from "node:module";
const require = createRequire(import.meta.url);

export default function (
  moduleName: string,
  dirname: string,
  absoluteRuntime: string | boolean,
) {
  if (absoluteRuntime === false) return moduleName;

  return resolveAbsoluteRuntime(
    moduleName,
    path.resolve(dirname, absoluteRuntime === true ? "." : absoluteRuntime),
  );
***REMOVED***

function resolveAbsoluteRuntime(moduleName: string, dirname: string) {
  try {
    return path
      .dirname(
        require.resolve(`${moduleName***REMOVED***/package.json`, { paths: [dirname] ***REMOVED***),
      )
      .replace(/\\/g, "/");
  ***REMOVED*** catch (err) {
    if (err.code !== "MODULE_NOT_FOUND") throw err;

    throw Object.assign(
      new Error(`Failed to resolve "${moduleName***REMOVED***" relative to "${dirname***REMOVED***"`),
      {
        code: "BABEL_RUNTIME_NOT_FOUND",
        runtime: moduleName,
        dirname,
      ***REMOVED***,
    );
  ***REMOVED***
***REMOVED***

export function resolveFSPath(path: string) {
  return require.resolve(path).replace(/\\/g, "/");
***REMOVED***
