// eslint/lib/cli-engine/formatters/html.js
/**
 * @fileoverview HTML reporter
 * @author Julian Laval
 */
"use strict";

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

const encodeHTML = (function() {
    const encodeHTMLRules = {
        "&": "&#38;",
        "<": "&#60;",
        ">": "&#62;",
        '"': "&#34;",
        "'": "&#39;"
    ***REMOVED***;
    const matchHTML = /[&<>"']/ug;

    return function(code) {
        return code
            ? code.toString().replace(matchHTML, m => encodeHTMLRules[m] || m)
            : "";
    ***REMOVED***;
***REMOVED***());

/**
 * Get the final HTML document.
 * @param {Object***REMOVED*** it data for the document.
 * @returns {string***REMOVED*** HTML document.
 */
function pageTemplate(it) {
    const { reportColor, reportSummary, date, results ***REMOVED*** = it;

    return `
<!DOCTYPE html>
<html>
    <head>
        <meta charset="UTF-8">
        <title>ESLint Report</title>
        <style>
            body {
                font-family:Arial, "Helvetica Neue", Helvetica, sans-serif;
                font-size:16px;
                font-weight:normal;
                margin:0;
                padding:0;
                color:#333
            ***REMOVED***
            #overview {
                padding:20px 30px
            ***REMOVED***
            td, th {
                padding:5px 10px
            ***REMOVED***
            h1 {
                margin:0
            ***REMOVED***
            table {
                margin:30px;
                width:calc(100% - 60px);
                max-width:1000px;
                border-radius:5px;
                border:1px solid #ddd;
                border-spacing:0px;
            ***REMOVED***
            th {
                font-weight:400;
                font-size:medium;
                text-align:left;
                cursor:pointer
            ***REMOVED***
            td.clr-1, td.clr-2, th span {
                font-weight:700
            ***REMOVED***
            th span {
                float:right;
                margin-left:20px
            ***REMOVED***
            th span:after {
                content:"";
                clear:both;
                display:block
            ***REMOVED***
            tr:last-child td {
                border-bottom:none
            ***REMOVED***
            tr td:first-child, tr td:last-child {
                color:#9da0a4
            ***REMOVED***
            #overview.bg-0, tr.bg-0 th {
                color:#468847;
                background:#dff0d8;
                border-bottom:1px solid #d6e9c6
            ***REMOVED***
            #overview.bg-1, tr.bg-1 th {
                color:#f0ad4e;
                background:#fcf8e3;
                border-bottom:1px solid #fbeed5
            ***REMOVED***
            #overview.bg-2, tr.bg-2 th {
                color:#b94a48;
                background:#f2dede;
                border-bottom:1px solid #eed3d7
            ***REMOVED***
            td {
                border-bottom:1px solid #ddd
            ***REMOVED***
            td.clr-1 {
                color:#f0ad4e
            ***REMOVED***
            td.clr-2 {
                color:#b94a48
            ***REMOVED***
            td a {
                color:#3a33d1;
                text-decoration:none
            ***REMOVED***
            td a:hover {
                color:#272296;
                text-decoration:underline
            ***REMOVED***
        </style>
    </head>
    <body>
        <div id="overview" class="bg-${reportColor***REMOVED***">
            <h1>ESLint Report</h1>
            <div>
                <span>${reportSummary***REMOVED***</span> - Generated on ${date***REMOVED***
            </div>
        </div>
        <table>
            <tbody>
                ${results***REMOVED***
            </tbody>
        </table>
        <script type="text/javascript">
            var groups = document.querySelectorAll("tr[data-group]");
            for (i = 0; i < groups.length; i++) {
                groups[i].addEventListener("click", function() {
                    var inGroup = document.getElementsByClassName(this.getAttribute("data-group"));
                    this.innerHTML = (this.innerHTML.indexOf("+") > -1) ? this.innerHTML.replace("+", "-") : this.innerHTML.replace("-", "+");
                    for (var j = 0; j < inGroup.length; j++) {
                        inGroup[j].style.display = (inGroup[j].style.display !== "none") ? "none" : "table-row";
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        </script>
    </body>
</html>
`.trimLeft();
***REMOVED***

/**
 * Given a word and a count, append an s if count is not one.
 * @param {string***REMOVED*** word A word in its singular form.
 * @param {int***REMOVED*** count A number controlling whether word should be pluralized.
 * @returns {string***REMOVED*** The original word with an s on the end if count is not one.
 */
function pluralize(word, count) {
    return (count === 1 ? word : `${word***REMOVED***s`);
***REMOVED***

/**
 * Renders text along the template of x problems (x errors, x warnings)
 * @param {string***REMOVED*** totalErrors Total errors
 * @param {string***REMOVED*** totalWarnings Total warnings
 * @returns {string***REMOVED*** The formatted string, pluralized where necessary
 */
function renderSummary(totalErrors, totalWarnings) {
    const totalProblems = totalErrors + totalWarnings;
    let renderedText = `${totalProblems***REMOVED*** ${pluralize("problem", totalProblems)***REMOVED***`;

    if (totalProblems !== 0) {
        renderedText += ` (${totalErrors***REMOVED*** ${pluralize("error", totalErrors)***REMOVED***, ${totalWarnings***REMOVED*** ${pluralize("warning", totalWarnings)***REMOVED***)`;
    ***REMOVED***
    return renderedText;
***REMOVED***

/**
 * Get the color based on whether there are errors/warnings...
 * @param {string***REMOVED*** totalErrors Total errors
 * @param {string***REMOVED*** totalWarnings Total warnings
 * @returns {int***REMOVED*** The color code (0 = green, 1 = yellow, 2 = red)
 */
function renderColor(totalErrors, totalWarnings) {
    if (totalErrors !== 0) {
        return 2;
    ***REMOVED***
    if (totalWarnings !== 0) {
        return 1;
    ***REMOVED***
    return 0;
***REMOVED***

/**
 * Get HTML (table row) describing a single message.
 * @param {Object***REMOVED*** it data for the message.
 * @returns {string***REMOVED*** HTML (table row) describing the message.
 */
function messageTemplate(it) {
    const {
        parentIndex,
        lineNumber,
        columnNumber,
        severityNumber,
        severityName,
        message,
        ruleUrl,
        ruleId
    ***REMOVED*** = it;

    return `
<tr style="display:none" class="f-${parentIndex***REMOVED***">
    <td>${lineNumber***REMOVED***:${columnNumber***REMOVED***</td>
    <td class="clr-${severityNumber***REMOVED***">${severityName***REMOVED***</td>
    <td>${encodeHTML(message)***REMOVED***</td>
    <td>
        <a href="${ruleUrl ? ruleUrl : ""***REMOVED***" target="_blank" rel="noopener noreferrer">${ruleId ? ruleId : ""***REMOVED***</a>
    </td>
</tr>
`.trimLeft();
***REMOVED***

/**
 * Get HTML (table rows) describing the messages.
 * @param {Array***REMOVED*** messages Messages.
 * @param {int***REMOVED*** parentIndex Index of the parent HTML row.
 * @param {Object***REMOVED*** rulesMeta Dictionary containing metadata for each rule executed by the analysis.
 * @returns {string***REMOVED*** HTML (table rows) describing the messages.
 */
function renderMessages(messages, parentIndex, rulesMeta) {

    /**
     * Get HTML (table row) describing a message.
     * @param {Object***REMOVED*** message Message.
     * @returns {string***REMOVED*** HTML (table row) describing a message.
     */
    return messages.map(message => {
        const lineNumber = message.line || 0;
        const columnNumber = message.column || 0;
        let ruleUrl;

        if (rulesMeta) {
            const meta = rulesMeta[message.ruleId];

            if (meta && meta.docs && meta.docs.url) {
                ruleUrl = meta.docs.url;
            ***REMOVED***
        ***REMOVED***

        return messageTemplate({
            parentIndex,
            lineNumber,
            columnNumber,
            severityNumber: message.severity,
            severityName: message.severity === 1 ? "Warning" : "Error",
            message: message.message,
            ruleId: message.ruleId,
            ruleUrl
        ***REMOVED***);
    ***REMOVED***).join("\n");
***REMOVED***

/**
 * Get HTML (table row) describing the result for a single file.
 * @param {Object***REMOVED*** it data for the file.
 * @returns {string***REMOVED*** HTML (table row) describing the result for the file.
 */
function resultTemplate(it) {
    const { color, index, filePath, summary ***REMOVED*** = it;

    return `
<tr class="bg-${color***REMOVED***" data-group="f-${index***REMOVED***">
    <th colspan="4">
        [+] ${encodeHTML(filePath)***REMOVED***
        <span>${encodeHTML(summary)***REMOVED***</span>
    </th>
</tr>
`.trimLeft();
***REMOVED***

// eslint-disable-next-line jsdoc/require-description
/**
 * @param {Array***REMOVED*** results Test results.
 * @param {Object***REMOVED*** rulesMeta Dictionary containing metadata for each rule executed by the analysis.
 * @returns {string***REMOVED*** HTML string describing the results.
 */
function renderResults(results, rulesMeta) {
    return results.map((result, index) => resultTemplate({
        index,
        color: renderColor(result.errorCount, result.warningCount),
        filePath: result.filePath,
        summary: renderSummary(result.errorCount, result.warningCount)
    ***REMOVED***) + renderMessages(result.messages, index, rulesMeta)).join("\n");
***REMOVED***

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

module.exports = function(results, data) {
    let totalErrors,
        totalWarnings;

    const metaData = data ? data.rulesMeta : {***REMOVED***;

    totalErrors = 0;
    totalWarnings = 0;

    // Iterate over results to get totals
    results.forEach(result => {
        totalErrors += result.errorCount;
        totalWarnings += result.warningCount;
    ***REMOVED***);

    return pageTemplate({
        date: new Date(),
        reportColor: renderColor(totalErrors, totalWarnings),
        reportSummary: renderSummary(totalErrors, totalWarnings),
        results: renderResults(results, metaData)
    ***REMOVED***);
***REMOVED***;
