// clean-css/lib/options/rounding-precision.js
var override = require('../utils/override');

var INTEGER_PATTERN = /^\d+$/;

var ALL_UNITS = ['*', 'all'];
var DEFAULT_PRECISION = 'off'; // all precision changes are disabled
var DIRECTIVES_SEPARATOR = ','; // e.g. *=5,px=3
var DIRECTIVE_VALUE_SEPARATOR = '='; // e.g. *=5

function roundingPrecisionFrom(source) {
  return override(defaults(DEFAULT_PRECISION), buildPrecisionFrom(source));
***REMOVED***

function defaults(value) {
  return {
    ch: value,
    cm: value,
    em: value,
    ex: value,
    in: value,
    mm: value,
    pc: value,
    pt: value,
    px: value,
    q: value,
    rem: value,
    vh: value,
    vmax: value,
    vmin: value,
    vw: value,
    '%': value
  ***REMOVED***;
***REMOVED***

function buildPrecisionFrom(source) {
  if (source === null || source === undefined) {
    return {***REMOVED***;
  ***REMOVED***

  if (typeof source == 'boolean') {
    return {***REMOVED***;
  ***REMOVED***

  if (typeof source == 'number' && source == -1) {
    return defaults(DEFAULT_PRECISION);
  ***REMOVED***

  if (typeof source == 'number') {
    return defaults(source);
  ***REMOVED***

  if (typeof source == 'string' && INTEGER_PATTERN.test(source)) {
    return defaults(parseInt(source));
  ***REMOVED***

  if (typeof source == 'string' && source == DEFAULT_PRECISION) {
    return defaults(DEFAULT_PRECISION);
  ***REMOVED***

  if (typeof source == 'object') {
    return source;
  ***REMOVED***

  return source
    .split(DIRECTIVES_SEPARATOR)
    .reduce(function(accumulator, directive) {
      var directiveParts = directive.split(DIRECTIVE_VALUE_SEPARATOR);
      var name = directiveParts[0];
      var value = parseInt(directiveParts[1]);

      if (Number.isNaN(value) || value == -1) {
        value = DEFAULT_PRECISION;
      ***REMOVED***

      if (ALL_UNITS.indexOf(name) > -1) {
        accumulator = override(accumulator, defaults(value));
      ***REMOVED*** else {
        accumulator[name] = value;
      ***REMOVED***

      return accumulator;
    ***REMOVED***, {***REMOVED***);
***REMOVED***

module.exports = {
  DEFAULT: DEFAULT_PRECISION,
  roundingPrecisionFrom: roundingPrecisionFrom
***REMOVED***;
