// fork-ts-checker-webpack-plugin/lib/profile/Performance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const perf_hooks_1 = require("perf_hooks");
function createPerformance() {
    let enabled = false;
    let timeOrigin;
    let marks;
    let measurements;
    function enable() {
        enabled = true;
        marks = new Map();
        measurements = new Map();
        timeOrigin = perf_hooks_1.performance.now();
    ***REMOVED***
    function disable() {
        enabled = false;
    ***REMOVED***
    function mark(name) {
        if (enabled) {
            marks.set(name, perf_hooks_1.performance.now());
        ***REMOVED***
    ***REMOVED***
    function measure(name, startMark, endMark) {
        if (enabled) {
            const start = (startMark && marks.get(startMark)) || timeOrigin;
            const end = (endMark && marks.get(endMark)) || perf_hooks_1.performance.now();
            measurements.set(name, (measurements.get(name) || 0) + (end - start));
        ***REMOVED***
    ***REMOVED***
    function markStart(name) {
        if (enabled) {
            mark(`${name***REMOVED*** start`);
        ***REMOVED***
    ***REMOVED***
    function markEnd(name) {
        if (enabled) {
            mark(`${name***REMOVED*** end`);
            measure(name, `${name***REMOVED*** start`, `${name***REMOVED*** end`);
        ***REMOVED***
    ***REMOVED***
    function formatName(name, width = 0) {
        return `${name***REMOVED***:`.padEnd(width);
    ***REMOVED***
    function formatDuration(duration, width = 0) {
        return `${(duration / 1000).toFixed(2)***REMOVED*** s`.padStart(width);
    ***REMOVED***
    function print() {
        if (enabled) {
            let nameWidth = 0;
            let durationWidth = 0;
            measurements.forEach((duration, name) => {
                nameWidth = Math.max(nameWidth, formatName(name).length);
                durationWidth = Math.max(durationWidth, formatDuration(duration).length);
            ***REMOVED***);
            measurements.forEach((duration, name) => {
                console.log(`${formatName(name, nameWidth)***REMOVED*** ${formatDuration(duration, durationWidth)***REMOVED***`);
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***
    return { enable, disable, mark, markStart, markEnd, measure, print ***REMOVED***;
***REMOVED***
exports.createPerformance = createPerformance;
