// workbox-build/src/lib/bundle.ts
/*
  Copyright 2019 Google LLC

  Use of this source code is governed by an MIT-style
  license that can be found in the LICENSE file or at
  https://opensource.org/licenses/MIT.
*/

import {babel***REMOVED*** from '@rollup/plugin-babel';
import {nodeResolve***REMOVED*** from '@rollup/plugin-node-resolve';
import {rollup, Plugin***REMOVED*** from 'rollup';
import {terser***REMOVED*** from 'rollup-plugin-terser';
import {writeFile***REMOVED*** from 'fs-extra';
import omt from '@surma/rollup-plugin-off-main-thread';
import presetEnv from '@babel/preset-env';
import replace from '@rollup/plugin-replace';
import tempy from 'tempy';
import upath from 'upath';

import {GeneratePartial, RequiredSWDestPartial***REMOVED*** from '../types';

interface NameAndContents {
  contents: string | Uint8Array;
  name: string;
***REMOVED***

export async function bundle({
  babelPresetEnvTargets,
  inlineWorkboxRuntime,
  mode,
  sourcemap,
  swDest,
  unbundledCode,
***REMOVED***: Omit<GeneratePartial, 'runtimeCaching'> &
  RequiredSWDestPartial & {unbundledCode: string***REMOVED***): Promise<
  Array<NameAndContents>
> {
  // We need to write this to the "real" file system, as Rollup won't read from
  // a custom file system.
  const {dir, base***REMOVED*** = upath.parse(swDest);

  const temporaryFile = tempy.file({name: base***REMOVED***);
  await writeFile(temporaryFile, unbundledCode);

  const plugins = [
    nodeResolve(),
    replace({
      // See https://github.com/GoogleChrome/workbox/issues/2769
      'preventAssignment': true,
      'process.env.NODE_ENV': JSON.stringify(mode),
    ***REMOVED***),
    babel({
      babelHelpers: 'bundled',
      // Disable the logic that checks for local Babel config files:
      // https://github.com/GoogleChrome/workbox/issues/2111
      babelrc: false,
      configFile: false,
      presets: [
        [
          presetEnv,
          {
            targets: {
              browsers: babelPresetEnvTargets,
            ***REMOVED***,
            loose: true,
          ***REMOVED***,
        ],
      ],
    ***REMOVED***),
  ];

  if (mode === 'production') {
    plugins.push(
      terser({
        mangle: {
          toplevel: true,
          properties: {
            regex: /(^_|_$)/,
          ***REMOVED***,
        ***REMOVED***,
      ***REMOVED***),
    );
  ***REMOVED***

  const rollupConfig: {
    input: string;
    manualChunks?: (id: string) => string | undefined;
    plugins: Array<Plugin>;
  ***REMOVED*** = {
    plugins,
    input: temporaryFile,
  ***REMOVED***;

  // Rollup will inline the runtime by default. If we don't want that, we need
  // to add in some additional config.
  if (!inlineWorkboxRuntime) {
    // No lint for omt(), library has no types.
    // eslint-disable-next-line  @typescript-eslint/no-unsafe-call
    rollupConfig.plugins.unshift(omt());
    rollupConfig.manualChunks = (id) => {
      return id.includes('workbox') ? 'workbox' : undefined;
    ***REMOVED***;
  ***REMOVED***

  const bundle = await rollup(rollupConfig);

  const {output***REMOVED*** = await bundle.generate({
    sourcemap,
    // Using an external Workbox runtime requires 'amd'.
    format: inlineWorkboxRuntime ? 'es' : 'amd',
  ***REMOVED***);

  const files: Array<NameAndContents> = [];
  for (const chunkOrAsset of output) {
    if (chunkOrAsset.type === 'asset') {
      files.push({
        name: chunkOrAsset.fileName,
        contents: chunkOrAsset.source,
      ***REMOVED***);
    ***REMOVED*** else {
      let code = chunkOrAsset.code;

      if (chunkOrAsset.map) {
        const sourceMapFile = chunkOrAsset.fileName + '.map';
        code += `//# sourceMappingURL=${sourceMapFile***REMOVED***\n`;

        files.push({
          name: sourceMapFile,
          contents: chunkOrAsset.map.toString(),
        ***REMOVED***);
      ***REMOVED***

      files.push({
        name: chunkOrAsset.fileName,
        contents: code,
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***

  // Make sure that if there was a directory portion included in swDest, it's
  // preprended to all of the generated files.
  return files.map((file) => {
    file.name = upath.format({
      dir,
      base: file.name,
      ext: '',
      name: '',
      root: '',
    ***REMOVED***);
    return file;
  ***REMOVED***);
***REMOVED***
