// workbox-build/src/lib/get-file-details.ts
/*
  Copyright 2021 Google LLC

  Use of this source code is governed by an MIT-style
  license that can be found in the LICENSE file or at
  https://opensource.org/licenses/MIT.
*/

import glob from 'glob';
import upath from 'upath';

import {errors***REMOVED*** from './errors';
import {getFileSize***REMOVED*** from './get-file-size';
import {getFileHash***REMOVED*** from './get-file-hash';

import {GlobPartial***REMOVED*** from '../types';

interface FileDetails {
  file: string;
  hash: string;
  size: number;
***REMOVED***

export function getFileDetails({
  globDirectory,
  globFollow,
  globIgnores,
  globPattern,
  globStrict,
***REMOVED***: Omit<GlobPartial, 'globDirectory' | 'globPatterns' | 'templatedURLs'> & {
  // This will only be called when globDirectory is not undefined.
  globDirectory: string;
  globPattern: string;
***REMOVED***): {
  globbedFileDetails: Array<FileDetails>;
  warning: string;
***REMOVED*** {
  let globbedFiles: Array<string>;
  let warning = '';

  try {
    globbedFiles = glob.sync(globPattern, {
      cwd: globDirectory,
      follow: globFollow,
      ignore: globIgnores,
      strict: globStrict,
    ***REMOVED***);
  ***REMOVED*** catch (err) {
    throw new Error(
      errors['unable-to-glob-files'] +
        ` '${err instanceof Error && err.message ? err.message : ''***REMOVED***'`,
    );
  ***REMOVED***

  if (globbedFiles.length === 0) {
    warning =
      errors['useless-glob-pattern'] +
      ' ' +
      JSON.stringify({globDirectory, globPattern, globIgnores***REMOVED***, null, 2);
  ***REMOVED***

  const globbedFileDetails: Array<FileDetails> = [];
  for (const file of globbedFiles) {
    const fullPath = upath.join(globDirectory, file);
    const fileSize = getFileSize(fullPath);
    if (fileSize !== null) {
      const fileHash = getFileHash(fullPath);
      globbedFileDetails.push({
        file: `${upath.relative(globDirectory, fullPath)***REMOVED***`,
        hash: fileHash,
        size: fileSize,
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***

  return {globbedFileDetails, warning***REMOVED***;
***REMOVED***
