// workbox-build/src/lib/replace-and-update-source-map.ts
/*
  Copyright 2019 Google LLC

  Use of this source code is governed by an MIT-style
  license that can be found in the LICENSE file or at
  https://opensource.org/licenses/MIT.
*/

import {RawSourceMap, SourceMapConsumer, SourceMapGenerator***REMOVED*** from 'source-map';

/**
 * Adapted from https://github.com/nsams/sourcemap-aware-replace, with modern
 * JavaScript updates, along with additional properties copied from originalMap.
 *
 * @param {Object***REMOVED*** options
 * @param {string***REMOVED*** options.jsFilename The name for the file whose contents
 * correspond to originalSource.
 * @param {Object***REMOVED*** options.originalMap The sourcemap for originalSource,
 * prior to any replacements.
 * @param {string***REMOVED*** options.originalSource The source code, prior to any
 * replacements.
 * @param {string***REMOVED*** options.replaceString A string to swap in for searchString.
 * @param {string***REMOVED*** options.searchString A string in originalSource to replace.
 * Only the first occurrence will be replaced.
 * @return {{source: string, map: string***REMOVED******REMOVED*** An object containing both
 * originalSource with the replacement applied, and the modified originalMap.
 *
 * @private
 */
export async function replaceAndUpdateSourceMap({
  jsFilename,
  originalMap,
  originalSource,
  replaceString,
  searchString,
***REMOVED***: {
  jsFilename: string;
  originalMap: RawSourceMap;
  originalSource: string;
  replaceString: string;
  searchString: string;
***REMOVED***): Promise<{map: string; source: string***REMOVED***> {
  const generator = new SourceMapGenerator({
    file: jsFilename,
  ***REMOVED***);

  const consumer = await new SourceMapConsumer(originalMap);

  let pos: number;
  let src = originalSource;
  const replacements: Array<{line: number; column: number***REMOVED***> = [];
  let lineNum = 0;
  let filePos = 0;

  const lines = src.split('\n');
  for (let line of lines) {
    lineNum++;
    let searchPos = 0;
    while ((pos = line.indexOf(searchString, searchPos)) !== -1) {
      src =
        src.substring(0, filePos + pos) +
        replaceString +
        src.substring(filePos + pos + searchString.length);
      line =
        line.substring(0, pos) +
        replaceString +
        line.substring(pos + searchString.length);
      replacements.push({line: lineNum, column: pos***REMOVED***);
      searchPos = pos + replaceString.length;
    ***REMOVED***
    filePos += line.length + 1;
  ***REMOVED***

  replacements.reverse();

  consumer.eachMapping((mapping) => {
    for (const replacement of replacements) {
      if (
        replacement.line === mapping.generatedLine &&
        mapping.generatedColumn > replacement.column
      ) {
        const offset = searchString.length - replaceString.length;
        mapping.generatedColumn -= offset;
      ***REMOVED***
    ***REMOVED***

    if (mapping.source) {
      const newMapping = {
        generated: {
          line: mapping.generatedLine,
          column: mapping.generatedColumn,
        ***REMOVED***,
        original: {
          line: mapping.originalLine,
          column: mapping.originalColumn,
        ***REMOVED***,
        source: mapping.source,
      ***REMOVED***;
      return generator.addMapping(newMapping);
    ***REMOVED***

    return mapping;
  ***REMOVED***);

  consumer.destroy();
  // JSON.parse returns any.
  // eslint-disable-next-line  @typescript-eslint/no-unsafe-assignment
  const updatedSourceMap: RawSourceMap = Object.assign(
    JSON.parse(generator.toString()),
    {
      names: originalMap.names,
      sourceRoot: originalMap.sourceRoot,
      sources: originalMap.sources,
      sourcesContent: originalMap.sourcesContent,
    ***REMOVED***,
  );

  return {
    map: JSON.stringify(updatedSourceMap),
    source: src,
  ***REMOVED***;
***REMOVED***
