// @babel/plugin-transform-modules-commonjs/lib/lazy.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.lazyImportsHook = void 0;
var _core = require("@babel/core");
var _helperModuleTransforms = require("@babel/helper-module-transforms");
const lazyImportsHook = lazy => ({
  name: `${"@babel/plugin-transform-modules-commonjs"***REMOVED***/lazy`,
  version: "7.27.1",
  getWrapperPayload(source, metadata) {
    if ((0, _helperModuleTransforms.isSideEffectImport)(metadata) || metadata.reexportAll) {
      return null;
    ***REMOVED***
    if (lazy === true) {
      return source.includes(".") ? null : "lazy/function";
    ***REMOVED***
    if (Array.isArray(lazy)) {
      return !lazy.includes(source) ? null : "lazy/function";
    ***REMOVED***
    if (typeof lazy === "function") {
      return lazy(source) ? "lazy/function" : null;
    ***REMOVED***
  ***REMOVED***,
  buildRequireWrapper(name, init, payload, referenced) {
    if (payload === "lazy/function") {
      if (!referenced) return false;
      return _core.template.statement.ast`
        function ${name***REMOVED***() {
          const data = ${init***REMOVED***;
          ${name***REMOVED*** = function(){ return data; ***REMOVED***;
          return data;
        ***REMOVED***
      `;
    ***REMOVED***
  ***REMOVED***,
  wrapReference(ref, payload) {
    if (payload === "lazy/function") return _core.types.callExpression(ref, []);
  ***REMOVED***
***REMOVED***);
exports.lazyImportsHook = lazyImportsHook;
