// @svgr/babel-plugin-transform-svg-component/lib/util.js
"use strict";

exports.__esModule = true;
exports.getExport = exports.getImport = exports.getInterface = exports.getProps = void 0;

function typeAnnotation(typeAnnotation) {
  return {
    type: 'TypeAnnotation',
    typeAnnotation
  ***REMOVED***;
***REMOVED***

function genericTypeAnnotation(id, typeParameters = null) {
  return {
    type: 'GenericTypeAnnotation',
    id,
    typeParameters
  ***REMOVED***;
***REMOVED***

function typeParameters(params) {
  return {
    type: 'TypeParameterInstantiation',
    params
  ***REMOVED***;
***REMOVED***

function qualifiedTypeIdentifier(qualification, id) {
  return {
    type: 'QualifiedTypeIdentifier',
    qualification,
    id
  ***REMOVED***;
***REMOVED***

function intersectionTypeAnnotation(types) {
  return {
    type: 'IntersectionTypeAnnotation',
    types
  ***REMOVED***;
***REMOVED***

function interfaceDeclaration(id, body) {
  return {
    type: 'InterfaceDeclaration',
    id,
    typeParameters: null,
    extends: [],
    implements: [],
    mixins: [],
    body
  ***REMOVED***;
***REMOVED***

function objectTypeAnnotation(properties) {
  return {
    type: 'ObjectTypeAnnotation',
    properties
  ***REMOVED***;
***REMOVED***

function objectTypeProperty(key, value, optional = false) {
  return {
    type: 'ObjectTypeProperty',
    key,
    static: false,
    proto: false,
    kind: 'init',
    method: false,
    value,
    variance: null,
    optional
  ***REMOVED***;
***REMOVED***

function addTypeAnotation(obj, typeAnnotation, opts) {
  if (!opts.typescript) return obj;
  return { ...obj,
    typeAnnotation
  ***REMOVED***;
***REMOVED***

function getSvgPropsTypeAnnotation(t, opts) {
  if (opts.native) {
    return t.genericTypeAnnotation(t.identifier('SvgProps'));
  ***REMOVED***

  return genericTypeAnnotation(qualifiedTypeIdentifier(t.identifier('React'), t.identifier('SVGProps')), typeParameters([genericTypeAnnotation(t.identifier('SVGSVGElement'))]));
***REMOVED***

const getProps = ({
  types: t
***REMOVED***, opts) => {
  const props = [];

  if (opts.titleProp) {
    props.push(t.objectProperty(t.identifier('title'), t.identifier('title'), false, true));
    props.push(t.objectProperty(t.identifier('titleId'), t.identifier('titleId'), false, true));
  ***REMOVED***

  if (opts.expandProps && props.length > 0) {
    props.push(t.restElement(t.identifier('props')));
  ***REMOVED***

  const propsArgument = props.length > 0 ? t.objectPattern(props) : t.identifier('props');
  let propsTypeAnnotation;

  if (props.length > 0) {
    propsTypeAnnotation = genericTypeAnnotation(t.identifier('SVGRProps'));

    if (opts.expandProps) {
      propsTypeAnnotation = intersectionTypeAnnotation([getSvgPropsTypeAnnotation(t, opts), propsTypeAnnotation]);
    ***REMOVED***
  ***REMOVED*** else {
    propsTypeAnnotation = opts.expandProps ? getSvgPropsTypeAnnotation(t, opts) : t.objectPattern([]);
  ***REMOVED***

  const typedPropsArgument = addTypeAnotation(propsArgument, typeAnnotation(propsTypeAnnotation), opts);
  const args = [];
  if (opts.expandProps || props.length > 0 || opts.ref) args.push(typedPropsArgument);

  if (opts.ref) {
    const refArgument = t.identifier(opts.typescript ? 'svgRef?' : 'svgRef');
    const typedRefArgument = addTypeAnotation(refArgument, typeAnnotation(genericTypeAnnotation(qualifiedTypeIdentifier(t.identifier('React'), t.identifier('Ref')), typeParameters([opts.native ? genericTypeAnnotation(qualifiedTypeIdentifier(t.identifier('React'), t.identifier('Component')), typeParameters([genericTypeAnnotation(t.identifier('SvgProps'))])) : genericTypeAnnotation(t.identifier('SVGSVGElement'))]))), opts);
    args.push(typedRefArgument);
  ***REMOVED***

  return args;
***REMOVED***;

exports.getProps = getProps;

const getInterface = ({
  types: t
***REMOVED***, opts) => {
  if (!opts.typescript) return null;
  const properties = [];

  if (opts.titleProp) {
    properties.push(objectTypeProperty(t.identifier('title'), t.identifier('string'), true));
    properties.push(objectTypeProperty(t.identifier('titleId'), t.identifier('string'), true));
  ***REMOVED***

  if (properties.length === 0) return null;
  return interfaceDeclaration(t.identifier('SVGRProps'), objectTypeAnnotation(properties));
***REMOVED***;

exports.getInterface = getInterface;

const getImport = ({
  types: t
***REMOVED***, opts) => {
  const importDeclarations = [t.importDeclaration([t.importNamespaceSpecifier(t.identifier('React'))], t.stringLiteral('react'))];

  if (opts.native) {
    if (opts.native.expo) {
      importDeclarations.push(t.importDeclaration([], t.stringLiteral('expo')));
    ***REMOVED*** else {
      const imports = [t.importDefaultSpecifier(t.identifier('Svg'))];

      if (opts.typescript && opts.expandProps) {
        imports.push(t.importSpecifier(t.identifier('SvgProps'), t.identifier('SvgProps')));
      ***REMOVED***

      importDeclarations.push(t.importDeclaration(imports, t.stringLiteral('react-native-svg')));
    ***REMOVED***
  ***REMOVED***

  return importDeclarations;
***REMOVED***;

exports.getImport = getImport;

const getExport = ({
  template
***REMOVED***, opts) => {
  let result = '';
  let exportName = opts.state.componentName;
  const plugins = ['jsx'];

  if (opts.typescript) {
    plugins.push('typescript');
  ***REMOVED***

  if (opts.ref) {
    const nextExportName = `ForwardRef`;
    result += `const ${nextExportName***REMOVED*** = React.forwardRef(${exportName***REMOVED***)\n\n`;
    exportName = nextExportName;
  ***REMOVED***

  if (opts.memo) {
    const nextExportName = `Memo${exportName***REMOVED***`;
    result += `const ${nextExportName***REMOVED*** = React.memo(${exportName***REMOVED***)\n\n`;
    exportName = nextExportName;
  ***REMOVED***

  if (opts.state.caller && opts.state.caller.previousExport) {
    result += `${opts.state.caller.previousExport***REMOVED***\n`;
    result += `export { ${exportName***REMOVED*** as ${opts.namedExport***REMOVED*** ***REMOVED***`;
    return template.ast(result, {
      plugins
    ***REMOVED***);
  ***REMOVED***

  result += `export default ${exportName***REMOVED***`;
  return template.ast(result, {
    plugins
  ***REMOVED***);
***REMOVED***;

exports.getExport = getExport;
