# import/no-empty-named-blocks

ðŸ”§ðŸ’¡ This rule is automatically fixable by the [`--fix` CLI option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix) and manually fixable by [editor suggestions](https://eslint.org/docs/latest/use/core-concepts#rule-suggestions).

<!-- end auto-generated rule header -->

Reports the use of empty named import blocks.

## Rule Details

### Valid

```js
import { mod ***REMOVED*** from 'mod'
import Default, { mod ***REMOVED*** from 'mod'
```

When using typescript

```js
import type { mod ***REMOVED*** from 'mod'
```

When using flow

```js
import typeof { mod ***REMOVED*** from 'mod'
```

### Invalid

```js
import {***REMOVED*** from 'mod'
import Default, {***REMOVED*** from 'mod'
```

When using typescript

```js
import type Default, {***REMOVED*** from 'mod'
import type {***REMOVED*** from 'mod'
```

When using flow

```js
import typeof {***REMOVED*** from 'mod'
import typeof Default, {***REMOVED*** from 'mod'
```
