// @babel/core/lib/config/validation/option-assertions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.access = access;
exports.assertArray = assertArray;
exports.assertAssumptions = assertAssumptions;
exports.assertBabelrcSearch = assertBabelrcSearch;
exports.assertBoolean = assertBoolean;
exports.assertCallerMetadata = assertCallerMetadata;
exports.assertCompact = assertCompact;
exports.assertConfigApplicableTest = assertConfigApplicableTest;
exports.assertConfigFileSearch = assertConfigFileSearch;
exports.assertFunction = assertFunction;
exports.assertIgnoreList = assertIgnoreList;
exports.assertInputSourceMap = assertInputSourceMap;
exports.assertObject = assertObject;
exports.assertPluginList = assertPluginList;
exports.assertRootMode = assertRootMode;
exports.assertSourceMaps = assertSourceMaps;
exports.assertSourceType = assertSourceType;
exports.assertString = assertString;
exports.assertTargets = assertTargets;
exports.msg = msg;
function _helperCompilationTargets() {
  const data = require("@babel/helper-compilation-targets");
  _helperCompilationTargets = function () {
    return data;
  ***REMOVED***;
  return data;
***REMOVED***
var _options = require("./options.js");
function msg(loc) {
  switch (loc.type) {
    case "root":
      return ``;
    case "env":
      return `${msg(loc.parent)***REMOVED***.env["${loc.name***REMOVED***"]`;
    case "overrides":
      return `${msg(loc.parent)***REMOVED***.overrides[${loc.index***REMOVED***]`;
    case "option":
      return `${msg(loc.parent)***REMOVED***.${loc.name***REMOVED***`;
    case "access":
      return `${msg(loc.parent)***REMOVED***[${JSON.stringify(loc.name)***REMOVED***]`;
    default:
      throw new Error(`Assertion failure: Unknown type ${loc.type***REMOVED***`);
  ***REMOVED***
***REMOVED***
function access(loc, name) {
  return {
    type: "access",
    name,
    parent: loc
  ***REMOVED***;
***REMOVED***
function assertRootMode(loc, value) {
  if (value !== undefined && value !== "root" && value !== "upward" && value !== "upward-optional") {
    throw new Error(`${msg(loc)***REMOVED*** must be a "root", "upward", "upward-optional" or undefined`);
  ***REMOVED***
  return value;
***REMOVED***
function assertSourceMaps(loc, value) {
  if (value !== undefined && typeof value !== "boolean" && value !== "inline" && value !== "both") {
    throw new Error(`${msg(loc)***REMOVED*** must be a boolean, "inline", "both", or undefined`);
  ***REMOVED***
  return value;
***REMOVED***
function assertCompact(loc, value) {
  if (value !== undefined && typeof value !== "boolean" && value !== "auto") {
    throw new Error(`${msg(loc)***REMOVED*** must be a boolean, "auto", or undefined`);
  ***REMOVED***
  return value;
***REMOVED***
function assertSourceType(loc, value) {
  if (value !== undefined && value !== "module" && value !== "script" && value !== "unambiguous") {
    throw new Error(`${msg(loc)***REMOVED*** must be "module", "script", "unambiguous", or undefined`);
  ***REMOVED***
  return value;
***REMOVED***
function assertCallerMetadata(loc, value) {
  const obj = assertObject(loc, value);
  if (obj) {
    if (typeof obj.name !== "string") {
      throw new Error(`${msg(loc)***REMOVED*** set but does not contain "name" property string`);
    ***REMOVED***
    for (const prop of Object.keys(obj)) {
      const propLoc = access(loc, prop);
      const value = obj[prop];
      if (value != null && typeof value !== "boolean" && typeof value !== "string" && typeof value !== "number") {
        throw new Error(`${msg(propLoc)***REMOVED*** must be null, undefined, a boolean, a string, or a number.`);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  return value;
***REMOVED***
function assertInputSourceMap(loc, value) {
  if (value !== undefined && typeof value !== "boolean" && (typeof value !== "object" || !value)) {
    throw new Error(`${msg(loc)***REMOVED*** must be a boolean, object, or undefined`);
  ***REMOVED***
  return value;
***REMOVED***
function assertString(loc, value) {
  if (value !== undefined && typeof value !== "string") {
    throw new Error(`${msg(loc)***REMOVED*** must be a string, or undefined`);
  ***REMOVED***
  return value;
***REMOVED***
function assertFunction(loc, value) {
  if (value !== undefined && typeof value !== "function") {
    throw new Error(`${msg(loc)***REMOVED*** must be a function, or undefined`);
  ***REMOVED***
  return value;
***REMOVED***
function assertBoolean(loc, value) {
  if (value !== undefined && typeof value !== "boolean") {
    throw new Error(`${msg(loc)***REMOVED*** must be a boolean, or undefined`);
  ***REMOVED***
  return value;
***REMOVED***
function assertObject(loc, value) {
  if (value !== undefined && (typeof value !== "object" || Array.isArray(value) || !value)) {
    throw new Error(`${msg(loc)***REMOVED*** must be an object, or undefined`);
  ***REMOVED***
  return value;
***REMOVED***
function assertArray(loc, value) {
  if (value != null && !Array.isArray(value)) {
    throw new Error(`${msg(loc)***REMOVED*** must be an array, or undefined`);
  ***REMOVED***
  return value;
***REMOVED***
function assertIgnoreList(loc, value) {
  const arr = assertArray(loc, value);
  arr == null || arr.forEach((item, i) => assertIgnoreItem(access(loc, i), item));
  return arr;
***REMOVED***
function assertIgnoreItem(loc, value) {
  if (typeof value !== "string" && typeof value !== "function" && !(value instanceof RegExp)) {
    throw new Error(`${msg(loc)***REMOVED*** must be an array of string/Function/RegExp values, or undefined`);
  ***REMOVED***
  return value;
***REMOVED***
function assertConfigApplicableTest(loc, value) {
  if (value === undefined) {
    return value;
  ***REMOVED***
  if (Array.isArray(value)) {
    value.forEach((item, i) => {
      if (!checkValidTest(item)) {
        throw new Error(`${msg(access(loc, i))***REMOVED*** must be a string/Function/RegExp.`);
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED*** else if (!checkValidTest(value)) {
    throw new Error(`${msg(loc)***REMOVED*** must be a string/Function/RegExp, or an array of those`);
  ***REMOVED***
  return value;
***REMOVED***
function checkValidTest(value) {
  return typeof value === "string" || typeof value === "function" || value instanceof RegExp;
***REMOVED***
function assertConfigFileSearch(loc, value) {
  if (value !== undefined && typeof value !== "boolean" && typeof value !== "string") {
    throw new Error(`${msg(loc)***REMOVED*** must be a undefined, a boolean, a string, ` + `got ${JSON.stringify(value)***REMOVED***`);
  ***REMOVED***
  return value;
***REMOVED***
function assertBabelrcSearch(loc, value) {
  if (value === undefined || typeof value === "boolean") {
    return value;
  ***REMOVED***
  if (Array.isArray(value)) {
    value.forEach((item, i) => {
      if (!checkValidTest(item)) {
        throw new Error(`${msg(access(loc, i))***REMOVED*** must be a string/Function/RegExp.`);
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED*** else if (!checkValidTest(value)) {
    throw new Error(`${msg(loc)***REMOVED*** must be a undefined, a boolean, a string/Function/RegExp ` + `or an array of those, got ${JSON.stringify(value)***REMOVED***`);
  ***REMOVED***
  return value;
***REMOVED***
function assertPluginList(loc, value) {
  const arr = assertArray(loc, value);
  if (arr) {
    arr.forEach((item, i) => assertPluginItem(access(loc, i), item));
  ***REMOVED***
  return arr;
***REMOVED***
function assertPluginItem(loc, value) {
  if (Array.isArray(value)) {
    if (value.length === 0) {
      throw new Error(`${msg(loc)***REMOVED*** must include an object`);
    ***REMOVED***
    if (value.length > 3) {
      throw new Error(`${msg(loc)***REMOVED*** may only be a two-tuple or three-tuple`);
    ***REMOVED***
    assertPluginTarget(access(loc, 0), value[0]);
    if (value.length > 1) {
      const opts = value[1];
      if (opts !== undefined && opts !== false && (typeof opts !== "object" || Array.isArray(opts) || opts === null)) {
        throw new Error(`${msg(access(loc, 1))***REMOVED*** must be an object, false, or undefined`);
      ***REMOVED***
    ***REMOVED***
    if (value.length === 3) {
      const name = value[2];
      if (name !== undefined && typeof name !== "string") {
        throw new Error(`${msg(access(loc, 2))***REMOVED*** must be a string, or undefined`);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED*** else {
    assertPluginTarget(loc, value);
  ***REMOVED***
  return value;
***REMOVED***
function assertPluginTarget(loc, value) {
  if ((typeof value !== "object" || !value) && typeof value !== "string" && typeof value !== "function") {
    throw new Error(`${msg(loc)***REMOVED*** must be a string, object, function`);
  ***REMOVED***
  return value;
***REMOVED***
function assertTargets(loc, value) {
  if ((0, _helperCompilationTargets().isBrowsersQueryValid)(value)) return value;
  if (typeof value !== "object" || !value || Array.isArray(value)) {
    throw new Error(`${msg(loc)***REMOVED*** must be a string, an array of strings or an object`);
  ***REMOVED***
  const browsersLoc = access(loc, "browsers");
  const esmodulesLoc = access(loc, "esmodules");
  assertBrowsersList(browsersLoc, value.browsers);
  assertBoolean(esmodulesLoc, value.esmodules);
  for (const key of Object.keys(value)) {
    const val = value[key];
    const subLoc = access(loc, key);
    if (key === "esmodules") assertBoolean(subLoc, val);else if (key === "browsers") assertBrowsersList(subLoc, val);else if (!hasOwnProperty.call(_helperCompilationTargets().TargetNames, key)) {
      const validTargets = Object.keys(_helperCompilationTargets().TargetNames).join(", ");
      throw new Error(`${msg(subLoc)***REMOVED*** is not a valid target. Supported targets are ${validTargets***REMOVED***`);
    ***REMOVED*** else assertBrowserVersion(subLoc, val);
  ***REMOVED***
  return value;
***REMOVED***
function assertBrowsersList(loc, value) {
  if (value !== undefined && !(0, _helperCompilationTargets().isBrowsersQueryValid)(value)) {
    throw new Error(`${msg(loc)***REMOVED*** must be undefined, a string or an array of strings`);
  ***REMOVED***
***REMOVED***
function assertBrowserVersion(loc, value) {
  if (typeof value === "number" && Math.round(value) === value) return;
  if (typeof value === "string") return;
  throw new Error(`${msg(loc)***REMOVED*** must be a string or an integer number`);
***REMOVED***
function assertAssumptions(loc, value) {
  if (value === undefined) return;
  if (typeof value !== "object" || value === null) {
    throw new Error(`${msg(loc)***REMOVED*** must be an object or undefined.`);
  ***REMOVED***
  let root = loc;
  do {
    root = root.parent;
  ***REMOVED*** while (root.type !== "root");
  const inPreset = root.source === "preset";
  for (const name of Object.keys(value)) {
    const subLoc = access(loc, name);
    if (!_options.assumptionsNames.has(name)) {
      throw new Error(`${msg(subLoc)***REMOVED*** is not a supported assumption.`);
    ***REMOVED***
    if (typeof value[name] !== "boolean") {
      throw new Error(`${msg(subLoc)***REMOVED*** must be a boolean.`);
    ***REMOVED***
    if (inPreset && value[name] === false) {
      throw new Error(`${msg(subLoc)***REMOVED*** cannot be set to 'false' inside presets.`);
    ***REMOVED***
  ***REMOVED***
  return value;
***REMOVED***
0 && 0;
