// @babel/core/lib/config/validation/plugins.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.validatePluginObject = validatePluginObject;
var _optionAssertions = require("./option-assertions.js");
const VALIDATORS = {
  name: _optionAssertions.assertString,
  manipulateOptions: _optionAssertions.assertFunction,
  pre: _optionAssertions.assertFunction,
  post: _optionAssertions.assertFunction,
  inherits: _optionAssertions.assertFunction,
  visitor: assertVisitorMap,
  parserOverride: _optionAssertions.assertFunction,
  generatorOverride: _optionAssertions.assertFunction
***REMOVED***;
function assertVisitorMap(loc, value) {
  const obj = (0, _optionAssertions.assertObject)(loc, value);
  if (obj) {
    Object.keys(obj).forEach(prop => {
      if (prop !== "_exploded" && prop !== "_verified") {
        assertVisitorHandler(prop, obj[prop]);
      ***REMOVED***
    ***REMOVED***);
    if (obj.enter || obj.exit) {
      throw new Error(`${(0, _optionAssertions.msg)(loc)***REMOVED*** cannot contain catch-all "enter" or "exit" handlers. Please target individual nodes.`);
    ***REMOVED***
  ***REMOVED***
  return obj;
***REMOVED***
function assertVisitorHandler(key, value) {
  if (value && typeof value === "object") {
    Object.keys(value).forEach(handler => {
      if (handler !== "enter" && handler !== "exit") {
        throw new Error(`.visitor["${key***REMOVED***"] may only have .enter and/or .exit handlers.`);
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED*** else if (typeof value !== "function") {
    throw new Error(`.visitor["${key***REMOVED***"] must be a function`);
  ***REMOVED***
***REMOVED***
function validatePluginObject(obj) {
  const rootPath = {
    type: "root",
    source: "plugin"
  ***REMOVED***;
  Object.keys(obj).forEach(key => {
    const validator = VALIDATORS[key];
    if (validator) {
      const optLoc = {
        type: "option",
        name: key,
        parent: rootPath
      ***REMOVED***;
      validator(optLoc, obj[key]);
    ***REMOVED*** else {
      const invalidPluginPropertyError = new Error(`.${key***REMOVED*** is not a valid Plugin property`);
      invalidPluginPropertyError.code = "BABEL_UNKNOWN_PLUGIN_PROPERTY";
      throw invalidPluginPropertyError;
    ***REMOVED***
  ***REMOVED***);
  return obj;
***REMOVED***
0 && 0;
