// jest-jasmine2/build/jasmine/Env.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
***REMOVED***);
exports.default = jasmineEnv;

var _assert = require('assert');

var _jestUtil = require('jest-util');

var _assertionErrorMessage = _interopRequireDefault(
  require('../assertionErrorMessage')
);

var _isError = _interopRequireDefault(require('../isError'));

var _queueRunner = _interopRequireDefault(require('../queueRunner'));

var _treeProcessor = _interopRequireDefault(require('../treeProcessor'));

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {default: obj***REMOVED***;
***REMOVED***

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    ***REMOVED***);
  ***REMOVED*** else {
    obj[key] = value;
  ***REMOVED***
  return obj;
***REMOVED***

function jasmineEnv(j$) {
  return class Env {
    constructor() {
      _defineProperty(this, 'specFilter', void 0);

      _defineProperty(this, 'catchExceptions', void 0);

      _defineProperty(this, 'throwOnExpectationFailure', void 0);

      _defineProperty(this, 'catchingExceptions', void 0);

      _defineProperty(this, 'topSuite', void 0);

      _defineProperty(this, 'fail', void 0);

      _defineProperty(this, 'pending', void 0);

      _defineProperty(this, 'afterAll', void 0);

      _defineProperty(this, 'fit', void 0);

      _defineProperty(this, 'throwingExpectationFailures', void 0);

      _defineProperty(this, 'randomizeTests', void 0);

      _defineProperty(this, 'randomTests', void 0);

      _defineProperty(this, 'seed', void 0);

      _defineProperty(this, 'execute', void 0);

      _defineProperty(this, 'fdescribe', void 0);

      _defineProperty(this, 'spyOn', void 0);

      _defineProperty(this, 'beforeEach', void 0);

      _defineProperty(this, 'afterEach', void 0);

      _defineProperty(this, 'clearReporters', void 0);

      _defineProperty(this, 'addReporter', void 0);

      _defineProperty(this, 'it', void 0);

      _defineProperty(this, 'xdescribe', void 0);

      _defineProperty(this, 'xit', void 0);

      _defineProperty(this, 'beforeAll', void 0);

      _defineProperty(this, 'todo', void 0);

      _defineProperty(this, 'provideFallbackReporter', void 0);

      _defineProperty(this, 'allowRespy', void 0);

      _defineProperty(this, 'describe', void 0);

      let totalSpecsDefined = 0;
      let catchExceptions = true;
      const realSetTimeout = global.setTimeout;
      const realClearTimeout = global.clearTimeout;
      const runnableResources = {***REMOVED***;
      const currentlyExecutingSuites = [];
      let currentSpec = null;
      let throwOnExpectationFailure = false;
      let random = false;
      let seed = null;
      let nextSpecId = 0;
      let nextSuiteId = 0;

      const getNextSpecId = function () {
        return 'spec' + nextSpecId++;
      ***REMOVED***;

      const getNextSuiteId = function () {
        return 'suite' + nextSuiteId++;
      ***REMOVED***;

      const topSuite = new j$.Suite({
        id: getNextSuiteId(),
        description: '',

        getTestPath() {
          return j$.testPath;
        ***REMOVED***
      ***REMOVED***);
      let currentDeclarationSuite = topSuite;

      const currentSuite = function () {
        return currentlyExecutingSuites[currentlyExecutingSuites.length - 1];
      ***REMOVED***;

      const currentRunnable = function () {
        return currentSpec || currentSuite();
      ***REMOVED***;

      const reporter = new j$.ReportDispatcher([
        'jasmineStarted',
        'jasmineDone',
        'suiteStarted',
        'suiteDone',
        'specStarted',
        'specDone'
      ]);

      this.specFilter = function () {
        return true;
      ***REMOVED***;

      const defaultResourcesForRunnable = function (id, _parentRunnableId) {
        const resources = {
          spies: []
        ***REMOVED***;
        runnableResources[id] = resources;
      ***REMOVED***;

      const clearResourcesForRunnable = function (id) {
        spyRegistry.clearSpies();
        delete runnableResources[id];
      ***REMOVED***;

      const beforeAndAfterFns = function (suite) {
        return function () {
          let afters = [];
          let befores = [];

          while (suite) {
            befores = befores.concat(suite.beforeFns);
            afters = afters.concat(suite.afterFns);
            suite = suite.parentSuite;
          ***REMOVED***

          return {
            befores: befores.reverse(),
            afters
          ***REMOVED***;
        ***REMOVED***;
      ***REMOVED***;

      const getSpecName = function (spec, suite) {
        const fullName = [spec.description];
        const suiteFullName = suite.getFullName();

        if (suiteFullName !== '') {
          fullName.unshift(suiteFullName);
        ***REMOVED***

        return fullName.join(' ');
      ***REMOVED***;

      this.catchExceptions = function (value) {
        catchExceptions = !!value;
        return catchExceptions;
      ***REMOVED***;

      this.catchingExceptions = function () {
        return catchExceptions;
      ***REMOVED***;

      this.throwOnExpectationFailure = function (value) {
        throwOnExpectationFailure = !!value;
      ***REMOVED***;

      this.throwingExpectationFailures = function () {
        return throwOnExpectationFailure;
      ***REMOVED***;

      this.randomizeTests = function (value) {
        random = !!value;
      ***REMOVED***;

      this.randomTests = function () {
        return random;
      ***REMOVED***;

      this.seed = function (value) {
        if (value) {
          seed = value;
        ***REMOVED***

        return seed;
      ***REMOVED***;

      const queueRunnerFactory = options => {
        options.clearTimeout = realClearTimeout;
        options.fail = this.fail;
        options.setTimeout = realSetTimeout;
        return (0, _queueRunner.default)(options);
      ***REMOVED***;

      this.topSuite = function () {
        return topSuite;
      ***REMOVED***;

      const uncaught = err => {
        if (currentSpec) {
          currentSpec.onException(err);
          currentSpec.cancel();
        ***REMOVED*** else {
          console.error('Unhandled error');
          console.error(err.stack);
        ***REMOVED***
      ***REMOVED***;

      let oldListenersException;
      let oldListenersRejection;

      const executionSetup = function () {
        // Need to ensure we are the only ones handling these exceptions.
        oldListenersException = process.listeners('uncaughtException').slice();
        oldListenersRejection = process.listeners('unhandledRejection').slice();
        j$.process.removeAllListeners('uncaughtException');
        j$.process.removeAllListeners('unhandledRejection');
        j$.process.on('uncaughtException', uncaught);
        j$.process.on('unhandledRejection', uncaught);
      ***REMOVED***;

      const executionTeardown = function () {
        j$.process.removeListener('uncaughtException', uncaught);
        j$.process.removeListener('unhandledRejection', uncaught); // restore previous exception handlers

        oldListenersException.forEach(listener => {
          j$.process.on('uncaughtException', listener);
        ***REMOVED***);
        oldListenersRejection.forEach(listener => {
          j$.process.on('unhandledRejection', listener);
        ***REMOVED***);
      ***REMOVED***;

      this.execute = async function (runnablesToRun, suiteTree = topSuite) {
        if (!runnablesToRun) {
          if (focusedRunnables.length) {
            runnablesToRun = focusedRunnables;
          ***REMOVED*** else {
            runnablesToRun = [suiteTree.id];
          ***REMOVED***
        ***REMOVED***

        if (currentlyExecutingSuites.length === 0) {
          executionSetup();
        ***REMOVED***

        const lastDeclarationSuite = currentDeclarationSuite;
        await (0, _treeProcessor.default)({
          nodeComplete(suite) {
            if (!suite.disabled) {
              clearResourcesForRunnable(suite.id);
            ***REMOVED***

            currentlyExecutingSuites.pop();

            if (suite === topSuite) {
              reporter.jasmineDone({
                failedExpectations: topSuite.result.failedExpectations
              ***REMOVED***);
            ***REMOVED*** else {
              reporter.suiteDone(suite.getResult());
            ***REMOVED***
          ***REMOVED***,

          nodeStart(suite) {
            currentlyExecutingSuites.push(suite);
            defaultResourcesForRunnable(
              suite.id,
              suite.parentSuite && suite.parentSuite.id
            );

            if (suite === topSuite) {
              reporter.jasmineStarted({
                totalSpecsDefined
              ***REMOVED***);
            ***REMOVED*** else {
              reporter.suiteStarted(suite.result);
            ***REMOVED***
          ***REMOVED***,

          queueRunnerFactory,
          runnableIds: runnablesToRun,
          tree: suiteTree
        ***REMOVED***);
        currentDeclarationSuite = lastDeclarationSuite;

        if (currentlyExecutingSuites.length === 0) {
          executionTeardown();
        ***REMOVED***
      ***REMOVED***;

      this.addReporter = function (reporterToAdd) {
        reporter.addReporter(reporterToAdd);
      ***REMOVED***;

      this.provideFallbackReporter = function (reporterToAdd) {
        reporter.provideFallbackReporter(reporterToAdd);
      ***REMOVED***;

      this.clearReporters = function () {
        reporter.clearReporters();
      ***REMOVED***;

      const spyRegistry = new j$.SpyRegistry({
        currentSpies() {
          if (!currentRunnable()) {
            throw new Error(
              'Spies must be created in a before function or a spec'
            );
          ***REMOVED***

          return runnableResources[currentRunnable().id].spies;
        ***REMOVED***
      ***REMOVED***);

      this.allowRespy = function (allow) {
        spyRegistry.allowRespy(allow);
      ***REMOVED***;

      this.spyOn = function (...args) {
        return spyRegistry.spyOn.apply(spyRegistry, args);
      ***REMOVED***;

      const suiteFactory = function (description) {
        const suite = new j$.Suite({
          id: getNextSuiteId(),
          description,
          parentSuite: currentDeclarationSuite,
          throwOnExpectationFailure,

          getTestPath() {
            return j$.testPath;
          ***REMOVED***
        ***REMOVED***);
        return suite;
      ***REMOVED***;

      this.describe = function (description, specDefinitions) {
        const suite = suiteFactory(description);

        if (specDefinitions === undefined) {
          throw new Error(
            'Missing second argument. It must be a callback function.'
          );
        ***REMOVED***

        if (typeof specDefinitions !== 'function') {
          throw new Error(
            `Invalid second argument, ${specDefinitions***REMOVED***. It must be a callback function.`
          );
        ***REMOVED***

        if (specDefinitions.length > 0) {
          throw new Error('describe does not expect any arguments');
        ***REMOVED***

        if (currentDeclarationSuite.markedPending) {
          suite.pend();
        ***REMOVED***

        if (currentDeclarationSuite.markedTodo) {
          // @ts-expect-error TODO Possible error: Suite does not have todo method
          suite.todo();
        ***REMOVED***

        addSpecsToSuite(suite, specDefinitions);
        return suite;
      ***REMOVED***;

      this.xdescribe = function (description, specDefinitions) {
        const suite = suiteFactory(description);
        suite.pend();
        addSpecsToSuite(suite, specDefinitions);
        return suite;
      ***REMOVED***;

      const focusedRunnables = [];

      this.fdescribe = function (description, specDefinitions) {
        const suite = suiteFactory(description);
        suite.isFocused = true;
        focusedRunnables.push(suite.id);
        unfocusAncestor();
        addSpecsToSuite(suite, specDefinitions);
        return suite;
      ***REMOVED***;

      const addSpecsToSuite = (suite, specDefinitions) => {
        const parentSuite = currentDeclarationSuite;
        parentSuite.addChild(suite);
        currentDeclarationSuite = suite;
        let declarationError = undefined;
        let describeReturnValue;

        try {
          describeReturnValue = specDefinitions.call(suite);
        ***REMOVED*** catch (e) {
          declarationError = e;
        ***REMOVED***

        if ((0, _jestUtil.isPromise)(describeReturnValue)) {
          declarationError = new Error(
            'Returning a Promise from "describe" is not supported. Tests must be defined synchronously.'
          );
        ***REMOVED*** else if (describeReturnValue !== undefined) {
          declarationError = new Error(
            'A "describe" callback must not return a value.'
          );
        ***REMOVED***

        if (declarationError) {
          this.it('encountered a declaration exception', () => {
            throw declarationError;
          ***REMOVED***);
        ***REMOVED***

        currentDeclarationSuite = parentSuite;
      ***REMOVED***;

      function findFocusedAncestor(suite) {
        while (suite) {
          if (suite.isFocused) {
            return suite.id;
          ***REMOVED***

          suite = suite.parentSuite;
        ***REMOVED***

        return null;
      ***REMOVED***

      function unfocusAncestor() {
        const focusedAncestor = findFocusedAncestor(currentDeclarationSuite);

        if (focusedAncestor) {
          for (let i = 0; i < focusedRunnables.length; i++) {
            if (focusedRunnables[i] === focusedAncestor) {
              focusedRunnables.splice(i, 1);
              break;
            ***REMOVED***
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***

      const specFactory = (description, fn, suite, timeout) => {
        totalSpecsDefined++;
        const spec = new j$.Spec({
          id: getNextSpecId(),
          beforeAndAfterFns: beforeAndAfterFns(suite),
          resultCallback: specResultCallback,

          getSpecName(spec) {
            return getSpecName(spec, suite);
          ***REMOVED***,

          getTestPath() {
            return j$.testPath;
          ***REMOVED***,

          onStart: specStarted,
          description,
          queueRunnerFactory,

          userContext() {
            return suite.clonedSharedUserContext();
          ***REMOVED***,

          queueableFn: {
            fn,

            timeout() {
              return timeout || j$._DEFAULT_TIMEOUT_INTERVAL;
            ***REMOVED***
          ***REMOVED***,
          throwOnExpectationFailure
        ***REMOVED***);

        if (!this.specFilter(spec)) {
          spec.disable();
        ***REMOVED***

        return spec;

        function specResultCallback(result) {
          clearResourcesForRunnable(spec.id);
          currentSpec = null;
          reporter.specDone(result);
        ***REMOVED***

        function specStarted(spec) {
          currentSpec = spec;
          defaultResourcesForRunnable(spec.id, suite.id);
          reporter.specStarted(spec.result);
        ***REMOVED***
      ***REMOVED***;

      this.it = function (description, fn, timeout) {
        if (typeof description !== 'string') {
          throw new Error(
            `Invalid first argument, ${description***REMOVED***. It must be a string.`
          );
        ***REMOVED***

        if (fn === undefined) {
          throw new Error(
            'Missing second argument. It must be a callback function. Perhaps you want to use `test.todo` for a test placeholder.'
          );
        ***REMOVED***

        if (typeof fn !== 'function') {
          throw new Error(
            `Invalid second argument, ${fn***REMOVED***. It must be a callback function.`
          );
        ***REMOVED***

        const spec = specFactory(
          description,
          fn,
          currentDeclarationSuite,
          timeout
        );

        if (currentDeclarationSuite.markedPending) {
          spec.pend();
        ***REMOVED*** // When a test is defined inside another, jasmine will not run it.
        // This check throws an error to warn the user about the edge-case.

        if (currentSpec !== null) {
          throw new Error(
            `Tests cannot be nested. Test "${spec.description***REMOVED***" cannot run because it is nested within "${currentSpec.description***REMOVED***".`
          );
        ***REMOVED***

        currentDeclarationSuite.addChild(spec);
        return spec;
      ***REMOVED***;

      this.xit = function (...args) {
        const spec = this.it.apply(this, args);
        spec.pend('Temporarily disabled with xit');
        return spec;
      ***REMOVED***;

      this.todo = function () {
        const description = arguments[0];

        if (arguments.length !== 1 || typeof description !== 'string') {
          throw new _jestUtil.ErrorWithStack(
            'Todo must be called with only a description.',
            this.todo
          );
        ***REMOVED***

        const spec = specFactory(
          description,
          () => {***REMOVED***,
          currentDeclarationSuite
        );

        if (currentDeclarationSuite.markedPending) {
          spec.pend();
        ***REMOVED*** else {
          spec.todo();
        ***REMOVED***

        currentDeclarationSuite.addChild(spec);
        return spec;
      ***REMOVED***;

      this.fit = function (description, fn, timeout) {
        const spec = specFactory(
          description,
          fn,
          currentDeclarationSuite,
          timeout
        );
        currentDeclarationSuite.addChild(spec);

        if (currentDeclarationSuite.markedPending) {
          spec.pend();
        ***REMOVED*** else {
          focusedRunnables.push(spec.id);
        ***REMOVED***

        unfocusAncestor();
        return spec;
      ***REMOVED***;

      this.beforeEach = function (beforeEachFunction, timeout) {
        currentDeclarationSuite.beforeEach({
          fn: beforeEachFunction,

          timeout() {
            return timeout || j$._DEFAULT_TIMEOUT_INTERVAL;
          ***REMOVED***
        ***REMOVED***);
      ***REMOVED***;

      this.beforeAll = function (beforeAllFunction, timeout) {
        currentDeclarationSuite.beforeAll({
          fn: beforeAllFunction,

          timeout() {
            return timeout || j$._DEFAULT_TIMEOUT_INTERVAL;
          ***REMOVED***
        ***REMOVED***);
      ***REMOVED***;

      this.afterEach = function (afterEachFunction, timeout) {
        currentDeclarationSuite.afterEach({
          fn: afterEachFunction,

          timeout() {
            return timeout || j$._DEFAULT_TIMEOUT_INTERVAL;
          ***REMOVED***
        ***REMOVED***);
      ***REMOVED***;

      this.afterAll = function (afterAllFunction, timeout) {
        currentDeclarationSuite.afterAll({
          fn: afterAllFunction,

          timeout() {
            return timeout || j$._DEFAULT_TIMEOUT_INTERVAL;
          ***REMOVED***
        ***REMOVED***);
      ***REMOVED***;

      this.pending = function (message) {
        let fullMessage = j$.Spec.pendingSpecExceptionMessage;

        if (message) {
          fullMessage += message;
        ***REMOVED***

        throw fullMessage;
      ***REMOVED***;

      this.fail = function (error) {
        let checkIsError;
        let message;

        if (
          error instanceof _assert.AssertionError ||
          (error && error.name === _assert.AssertionError.name)
        ) {
          checkIsError = false; // @ts-expect-error TODO Possible error: j$.Spec does not have expand property

          message = (0, _assertionErrorMessage.default)(error, {
            expand: j$.Spec.expand
          ***REMOVED***);
        ***REMOVED*** else {
          const check = (0, _isError.default)(error);
          checkIsError = check.isError;
          message = check.message;
        ***REMOVED***

        const errorAsErrorObject = checkIsError ? error : new Error(message);
        const runnable = currentRunnable();

        if (!runnable) {
          errorAsErrorObject.message =
            'Caught error after test environment was torn down\n\n' +
            errorAsErrorObject.message;
          throw errorAsErrorObject;
        ***REMOVED***

        runnable.addExpectationResult(false, {
          matcherName: '',
          passed: false,
          expected: '',
          actual: '',
          message,
          error: errorAsErrorObject
        ***REMOVED***);
      ***REMOVED***;
    ***REMOVED***
  ***REMOVED***;
***REMOVED***
