// jest-jasmine2/build/jasmine/Spec.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
***REMOVED***);
exports.default = void 0;

var _assert = require('assert');

var _ExpectationFailed = _interopRequireDefault(
  require('../ExpectationFailed')
);

var _assertionErrorMessage = _interopRequireDefault(
  require('../assertionErrorMessage')
);

var _expectationResultFactory = _interopRequireDefault(
  require('../expectationResultFactory')
);

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {default: obj***REMOVED***;
***REMOVED***

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    ***REMOVED***);
  ***REMOVED*** else {
    obj[key] = value;
  ***REMOVED***
  return obj;
***REMOVED***

class Spec {
  static isPendingSpecException(e) {
    return !!(
      e &&
      e.toString &&
      e.toString().indexOf(Spec.pendingSpecExceptionMessage) !== -1
    );
  ***REMOVED***

  constructor(attrs) {
    _defineProperty(this, 'id', void 0);

    _defineProperty(this, 'description', void 0);

    _defineProperty(this, 'resultCallback', void 0);

    _defineProperty(this, 'queueableFn', void 0);

    _defineProperty(this, 'beforeAndAfterFns', void 0);

    _defineProperty(this, 'userContext', void 0);

    _defineProperty(this, 'onStart', void 0);

    _defineProperty(this, 'getSpecName', void 0);

    _defineProperty(this, 'queueRunnerFactory', void 0);

    _defineProperty(this, 'throwOnExpectationFailure', void 0);

    _defineProperty(this, 'initError', void 0);

    _defineProperty(this, 'result', void 0);

    _defineProperty(this, 'disabled', void 0);

    _defineProperty(this, 'currentRun', void 0);

    _defineProperty(this, 'markedTodo', void 0);

    _defineProperty(this, 'markedPending', void 0);

    _defineProperty(this, 'expand', void 0);

    this.resultCallback = attrs.resultCallback || function () {***REMOVED***;

    this.id = attrs.id;
    this.description = attrs.description || '';
    this.queueableFn = attrs.queueableFn;

    this.beforeAndAfterFns =
      attrs.beforeAndAfterFns ||
      function () {
        return {
          befores: [],
          afters: []
        ***REMOVED***;
      ***REMOVED***;

    this.userContext =
      attrs.userContext ||
      function () {
        return {***REMOVED***;
      ***REMOVED***;

    this.onStart = attrs.onStart || function () {***REMOVED***;

    this.getSpecName =
      attrs.getSpecName ||
      function () {
        return '';
      ***REMOVED***;

    this.queueRunnerFactory = attrs.queueRunnerFactory || function () {***REMOVED***;

    this.throwOnExpectationFailure = !!attrs.throwOnExpectationFailure;
    this.initError = new Error();
    this.initError.name = ''; // Without this line v8 stores references to all closures
    // in the stack in the Error object. This line stringifies the stack
    // property to allow garbage-collecting objects on the stack
    // https://crbug.com/v8/7142

    this.initError.stack = this.initError.stack;
    this.queueableFn.initError = this.initError; // @ts-expect-error

    this.result = {
      id: this.id,
      description: this.description,
      fullName: this.getFullName(),
      failedExpectations: [],
      passedExpectations: [],
      pendingReason: '',
      testPath: attrs.getTestPath()
    ***REMOVED***;
  ***REMOVED***

  addExpectationResult(passed, data, isError) {
    const expectationResult = (0, _expectationResultFactory.default)(
      data,
      this.initError
    );

    if (passed) {
      this.result.passedExpectations.push(expectationResult);
    ***REMOVED*** else {
      this.result.failedExpectations.push(expectationResult);

      if (this.throwOnExpectationFailure && !isError) {
        throw new _ExpectationFailed.default();
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  execute(onComplete, enabled) {
    const self = this;
    this.onStart(this);

    if (
      !this.isExecutable() ||
      this.markedPending ||
      this.markedTodo ||
      enabled === false
    ) {
      complete(enabled);
      return;
    ***REMOVED***

    const fns = this.beforeAndAfterFns();
    const allFns = fns.befores.concat(this.queueableFn).concat(fns.afters);
    this.currentRun = this.queueRunnerFactory({
      queueableFns: allFns,

      onException() {
        // @ts-expect-error
        self.onException.apply(self, arguments);
      ***REMOVED***,

      userContext: this.userContext(),
      setTimeout,
      clearTimeout,
      fail: () => {***REMOVED***
    ***REMOVED***);
    this.currentRun.then(() => complete(true));

    function complete(enabledAgain) {
      self.result.status = self.status(enabledAgain);
      self.resultCallback(self.result);

      if (onComplete) {
        onComplete();
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  cancel() {
    if (this.currentRun) {
      this.currentRun.cancel();
    ***REMOVED***
  ***REMOVED***

  onException(error) {
    if (Spec.isPendingSpecException(error)) {
      this.pend(extractCustomPendingMessage(error));
      return;
    ***REMOVED***

    if (error instanceof _ExpectationFailed.default) {
      return;
    ***REMOVED***

    this.addExpectationResult(
      false,
      {
        matcherName: '',
        passed: false,
        expected: '',
        actual: '',
        error: this.isAssertionError(error)
          ? (0, _assertionErrorMessage.default)(error, {
              expand: this.expand
            ***REMOVED***)
          : error
      ***REMOVED***,
      true
    );
  ***REMOVED***

  disable() {
    this.disabled = true;
  ***REMOVED***

  pend(message) {
    this.markedPending = true;

    if (message) {
      this.result.pendingReason = message;
    ***REMOVED***
  ***REMOVED***

  todo() {
    this.markedTodo = true;
  ***REMOVED***

  getResult() {
    this.result.status = this.status();
    return this.result;
  ***REMOVED***

  status(enabled) {
    if (this.disabled || enabled === false) {
      return 'disabled';
    ***REMOVED***

    if (this.markedTodo) {
      return 'todo';
    ***REMOVED***

    if (this.markedPending) {
      return 'pending';
    ***REMOVED***

    if (this.result.failedExpectations.length > 0) {
      return 'failed';
    ***REMOVED*** else {
      return 'passed';
    ***REMOVED***
  ***REMOVED***

  isExecutable() {
    return !this.disabled;
  ***REMOVED***

  getFullName() {
    return this.getSpecName(this);
  ***REMOVED***

  isAssertionError(error) {
    return (
      error instanceof _assert.AssertionError ||
      (error && error.name === _assert.AssertionError.name)
    );
  ***REMOVED***
***REMOVED***

exports.default = Spec;

_defineProperty(Spec, 'pendingSpecExceptionMessage', void 0);

Spec.pendingSpecExceptionMessage = '=> marked Pending';

const extractCustomPendingMessage = function (e) {
  const fullMessage = e.toString();
  const boilerplateStart = fullMessage.indexOf(
    Spec.pendingSpecExceptionMessage
  );
  const boilerplateEnd =
    boilerplateStart + Spec.pendingSpecExceptionMessage.length;
  return fullMessage.substr(boilerplateEnd);
***REMOVED***;
