// get-stream/index.js
'use strict';
const {constants: BufferConstants***REMOVED*** = require('buffer');
const stream = require('stream');
const {promisify***REMOVED*** = require('util');
const bufferStream = require('./buffer-stream');

const streamPipelinePromisified = promisify(stream.pipeline);

class MaxBufferError extends Error {
	constructor() {
		super('maxBuffer exceeded');
		this.name = 'MaxBufferError';
	***REMOVED***
***REMOVED***

async function getStream(inputStream, options) {
	if (!inputStream) {
		throw new Error('Expected a stream');
	***REMOVED***

	options = {
		maxBuffer: Infinity,
		...options
	***REMOVED***;

	const {maxBuffer***REMOVED*** = options;
	const stream = bufferStream(options);

	await new Promise((resolve, reject) => {
		const rejectPromise = error => {
			// Don't retrieve an oversized buffer.
			if (error && stream.getBufferedLength() <= BufferConstants.MAX_LENGTH) {
				error.bufferedData = stream.getBufferedValue();
			***REMOVED***

			reject(error);
		***REMOVED***;

		(async () => {
			try {
				await streamPipelinePromisified(inputStream, stream);
				resolve();
			***REMOVED*** catch (error) {
				rejectPromise(error);
			***REMOVED***
		***REMOVED***)();

		stream.on('data', () => {
			if (stream.getBufferedLength() > maxBuffer) {
				rejectPromise(new MaxBufferError());
			***REMOVED***
		***REMOVED***);
	***REMOVED***);

	return stream.getBufferedValue();
***REMOVED***

module.exports = getStream;
module.exports.buffer = (stream, options) => getStream(stream, {...options, encoding: 'buffer'***REMOVED***);
module.exports.array = (stream, options) => getStream(stream, {...options, array: true***REMOVED***);
module.exports.MaxBufferError = MaxBufferError;
