// deepmerge/dist/umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global = global || self, global.deepmerge = factory());
***REMOVED***(this, function () { 'use strict';

	var isMergeableObject = function isMergeableObject(value) {
		return isNonNullObject(value)
			&& !isSpecial(value)
	***REMOVED***;

	function isNonNullObject(value) {
		return !!value && typeof value === 'object'
	***REMOVED***

	function isSpecial(value) {
		var stringValue = Object.prototype.toString.call(value);

		return stringValue === '[object RegExp]'
			|| stringValue === '[object Date]'
			|| isReactElement(value)
	***REMOVED***

	// see https://github.com/facebook/react/blob/b5ac963fb791d1298e7f396236383bc955f916c1/src/isomorphic/classic/element/ReactElement.js#L21-L25
	var canUseSymbol = typeof Symbol === 'function' && Symbol.for;
	var REACT_ELEMENT_TYPE = canUseSymbol ? Symbol.for('react.element') : 0xeac7;

	function isReactElement(value) {
		return value.$$typeof === REACT_ELEMENT_TYPE
	***REMOVED***

	function emptyTarget(val) {
		return Array.isArray(val) ? [] : {***REMOVED***
	***REMOVED***

	function cloneUnlessOtherwiseSpecified(value, options) {
		return (options.clone !== false && options.isMergeableObject(value))
			? deepmerge(emptyTarget(value), value, options)
			: value
	***REMOVED***

	function defaultArrayMerge(target, source, options) {
		return target.concat(source).map(function(element) {
			return cloneUnlessOtherwiseSpecified(element, options)
		***REMOVED***)
	***REMOVED***

	function getMergeFunction(key, options) {
		if (!options.customMerge) {
			return deepmerge
		***REMOVED***
		var customMerge = options.customMerge(key);
		return typeof customMerge === 'function' ? customMerge : deepmerge
	***REMOVED***

	function getEnumerableOwnPropertySymbols(target) {
		return Object.getOwnPropertySymbols
			? Object.getOwnPropertySymbols(target).filter(function(symbol) {
				return Object.propertyIsEnumerable.call(target, symbol)
			***REMOVED***)
			: []
	***REMOVED***

	function getKeys(target) {
		return Object.keys(target).concat(getEnumerableOwnPropertySymbols(target))
	***REMOVED***

	function propertyIsOnObject(object, property) {
		try {
			return property in object
		***REMOVED*** catch(_) {
			return false
		***REMOVED***
	***REMOVED***

	// Protects from prototype poisoning and unexpected merging up the prototype chain.
	function propertyIsUnsafe(target, key) {
		return propertyIsOnObject(target, key) // Properties are safe to merge if they don't exist in the target yet,
			&& !(Object.hasOwnProperty.call(target, key) // unsafe if they exist up the prototype chain,
				&& Object.propertyIsEnumerable.call(target, key)) // and also unsafe if they're nonenumerable.
	***REMOVED***

	function mergeObject(target, source, options) {
		var destination = {***REMOVED***;
		if (options.isMergeableObject(target)) {
			getKeys(target).forEach(function(key) {
				destination[key] = cloneUnlessOtherwiseSpecified(target[key], options);
			***REMOVED***);
		***REMOVED***
		getKeys(source).forEach(function(key) {
			if (propertyIsUnsafe(target, key)) {
				return
			***REMOVED***

			if (propertyIsOnObject(target, key) && options.isMergeableObject(source[key])) {
				destination[key] = getMergeFunction(key, options)(target[key], source[key], options);
			***REMOVED*** else {
				destination[key] = cloneUnlessOtherwiseSpecified(source[key], options);
			***REMOVED***
		***REMOVED***);
		return destination
	***REMOVED***

	function deepmerge(target, source, options) {
		options = options || {***REMOVED***;
		options.arrayMerge = options.arrayMerge || defaultArrayMerge;
		options.isMergeableObject = options.isMergeableObject || isMergeableObject;
		// cloneUnlessOtherwiseSpecified is added to `options` so that custom arrayMerge()
		// implementations can use it. The caller may not replace it.
		options.cloneUnlessOtherwiseSpecified = cloneUnlessOtherwiseSpecified;

		var sourceIsArray = Array.isArray(source);
		var targetIsArray = Array.isArray(target);
		var sourceAndTargetTypesMatch = sourceIsArray === targetIsArray;

		if (!sourceAndTargetTypesMatch) {
			return cloneUnlessOtherwiseSpecified(source, options)
		***REMOVED*** else if (sourceIsArray) {
			return options.arrayMerge(target, source, options)
		***REMOVED*** else {
			return mergeObject(target, source, options)
		***REMOVED***
	***REMOVED***

	deepmerge.all = function deepmergeAll(array, options) {
		if (!Array.isArray(array)) {
			throw new Error('first argument should be an array')
		***REMOVED***

		return array.reduce(function(prev, next) {
			return deepmerge(prev, next, options)
		***REMOVED***, {***REMOVED***)
	***REMOVED***;

	var deepmerge_1 = deepmerge;

	return deepmerge_1;

***REMOVED***));
