// @typescript-eslint/eslint-plugin/dist/rules/ban-types.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; ***REMOVED*** ***REMOVED***;
    ***REMOVED***
    Object.defineProperty(o, k2, desc);
***REMOVED***) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
***REMOVED***));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v ***REMOVED***);
***REMOVED***) : function(o, v) {
    o["default"] = v;
***REMOVED***);
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {***REMOVED***;
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.TYPE_KEYWORDS = void 0;
const utils_1 = require("@typescript-eslint/utils");
const util = __importStar(require("../util"));
function removeSpaces(str) {
    return str.replace(/\s/g, '');
***REMOVED***
function stringifyNode(node, sourceCode) {
    return removeSpaces(sourceCode.getText(node));
***REMOVED***
function getCustomMessage(bannedType) {
    if (bannedType == null) {
        return '';
    ***REMOVED***
    if (typeof bannedType === 'string') {
        return ` ${bannedType***REMOVED***`;
    ***REMOVED***
    if (bannedType.message) {
        return ` ${bannedType.message***REMOVED***`;
    ***REMOVED***
    return '';
***REMOVED***
const defaultTypes = {
    String: {
        message: 'Use string instead',
        fixWith: 'string',
    ***REMOVED***,
    Boolean: {
        message: 'Use boolean instead',
        fixWith: 'boolean',
    ***REMOVED***,
    Number: {
        message: 'Use number instead',
        fixWith: 'number',
    ***REMOVED***,
    Symbol: {
        message: 'Use symbol instead',
        fixWith: 'symbol',
    ***REMOVED***,
    BigInt: {
        message: 'Use bigint instead',
        fixWith: 'bigint',
    ***REMOVED***,
    Function: {
        message: [
            'The `Function` type accepts any function-like value.',
            'It provides no type safety when calling the function, which can be a common source of bugs.',
            'It also accepts things like class declarations, which will throw at runtime as they will not be called with `new`.',
            'If you are expecting the function to accept certain arguments, you should explicitly define the function shape.',
        ].join('\n'),
    ***REMOVED***,
    // object typing
    Object: {
        message: [
            'The `Object` type actually means "any non-nullish value", so it is marginally better than `unknown`.',
            '- If you want a type meaning "any object", you probably want `object` instead.',
            '- If you want a type meaning "any value", you probably want `unknown` instead.',
            '- If you really want a type meaning "any non-nullish value", you probably want `NonNullable<unknown>` instead.',
        ].join('\n'),
        suggest: ['object', 'unknown', 'NonNullable<unknown>'],
    ***REMOVED***,
    '{***REMOVED***': {
        message: [
            '`{***REMOVED***` actually means "any non-nullish value".',
            '- If you want a type meaning "any object", you probably want `object` instead.',
            '- If you want a type meaning "any value", you probably want `unknown` instead.',
            '- If you want a type meaning "empty object", you probably want `Record<string, never>` instead.',
            '- If you really want a type meaning "any non-nullish value", you probably want `NonNullable<unknown>` instead.',
        ].join('\n'),
        suggest: [
            'object',
            'unknown',
            'Record<string, never>',
            'NonNullable<unknown>',
        ],
    ***REMOVED***,
***REMOVED***;
exports.TYPE_KEYWORDS = {
    bigint: utils_1.AST_NODE_TYPES.TSBigIntKeyword,
    boolean: utils_1.AST_NODE_TYPES.TSBooleanKeyword,
    never: utils_1.AST_NODE_TYPES.TSNeverKeyword,
    null: utils_1.AST_NODE_TYPES.TSNullKeyword,
    number: utils_1.AST_NODE_TYPES.TSNumberKeyword,
    object: utils_1.AST_NODE_TYPES.TSObjectKeyword,
    string: utils_1.AST_NODE_TYPES.TSStringKeyword,
    symbol: utils_1.AST_NODE_TYPES.TSSymbolKeyword,
    undefined: utils_1.AST_NODE_TYPES.TSUndefinedKeyword,
    unknown: utils_1.AST_NODE_TYPES.TSUnknownKeyword,
    void: utils_1.AST_NODE_TYPES.TSVoidKeyword,
***REMOVED***;
exports.default = util.createRule({
    name: 'ban-types',
    meta: {
        type: 'suggestion',
        docs: {
            description: 'Disallow certain types',
            recommended: 'error',
        ***REMOVED***,
        fixable: 'code',
        hasSuggestions: true,
        messages: {
            bannedTypeMessage: "Don't use `{{name***REMOVED******REMOVED***` as a type.{{customMessage***REMOVED******REMOVED***",
            bannedTypeReplacement: 'Replace `{{name***REMOVED******REMOVED***` with `{{replacement***REMOVED******REMOVED***`.',
        ***REMOVED***,
        schema: [
            {
                type: 'object',
                properties: {
                    types: {
                        type: 'object',
                        additionalProperties: {
                            oneOf: [
                                { type: 'null' ***REMOVED***,
                                { type: 'boolean' ***REMOVED***,
                                { type: 'string' ***REMOVED***,
                                {
                                    type: 'object',
                                    properties: {
                                        message: { type: 'string' ***REMOVED***,
                                        fixWith: { type: 'string' ***REMOVED***,
                                        suggest: {
                                            type: 'array',
                                            items: { type: 'string' ***REMOVED***,
                                        ***REMOVED***,
                                    ***REMOVED***,
                                    additionalProperties: false,
                                ***REMOVED***,
                            ],
                        ***REMOVED***,
                    ***REMOVED***,
                    extendDefaults: {
                        type: 'boolean',
                    ***REMOVED***,
                ***REMOVED***,
                additionalProperties: false,
            ***REMOVED***,
        ],
    ***REMOVED***,
    defaultOptions: [{***REMOVED***],
    create(context, [options]) {
        var _a, _b;
        const extendDefaults = (_a = options.extendDefaults) !== null && _a !== void 0 ? _a : true;
        const customTypes = (_b = options.types) !== null && _b !== void 0 ? _b : {***REMOVED***;
        const types = Object.assign({***REMOVED***, extendDefaults ? defaultTypes : {***REMOVED***, customTypes);
        const bannedTypes = new Map(Object.entries(types).map(([type, data]) => [removeSpaces(type), data]));
        function checkBannedTypes(typeNode, name = stringifyNode(typeNode, context.getSourceCode())) {
            const bannedType = bannedTypes.get(name);
            if (bannedType === undefined || bannedType === false) {
                return;
            ***REMOVED***
            const customMessage = getCustomMessage(bannedType);
            const fixWith = bannedType && typeof bannedType === 'object' && bannedType.fixWith;
            const suggest = bannedType && typeof bannedType === 'object'
                ? bannedType.suggest
                : undefined;
            context.report({
                node: typeNode,
                messageId: 'bannedTypeMessage',
                data: {
                    name,
                    customMessage,
                ***REMOVED***,
                fix: fixWith
                    ? (fixer) => fixer.replaceText(typeNode, fixWith)
                    : null,
                suggest: suggest === null || suggest === void 0 ? void 0 : suggest.map(replacement => ({
                    messageId: 'bannedTypeReplacement',
                    data: {
                        name,
                        replacement,
                    ***REMOVED***,
                    fix: (fixer) => fixer.replaceText(typeNode, replacement),
                ***REMOVED***)),
            ***REMOVED***);
        ***REMOVED***
        const keywordSelectors = util.objectReduceKey(exports.TYPE_KEYWORDS, (acc, keyword) => {
            if (bannedTypes.has(keyword)) {
                acc[exports.TYPE_KEYWORDS[keyword]] = (node) => checkBannedTypes(node, keyword);
            ***REMOVED***
            return acc;
        ***REMOVED***, {***REMOVED***);
        return Object.assign(Object.assign({***REMOVED***, keywordSelectors), { TSTypeLiteral(node) {
                if (node.members.length) {
                    return;
                ***REMOVED***
                checkBannedTypes(node);
            ***REMOVED***,
            TSTupleType(node) {
                if (node.elementTypes.length === 0) {
                    checkBannedTypes(node);
                ***REMOVED***
            ***REMOVED***,
            TSTypeReference(node) {
                checkBannedTypes(node.typeName);
                if (node.typeParameters) {
                    checkBannedTypes(node);
                ***REMOVED***
            ***REMOVED***,
            TSInterfaceHeritage(node) {
                checkBannedTypes(node);
            ***REMOVED***,
            TSClassImplements(node) {
                checkBannedTypes(node);
            ***REMOVED*** ***REMOVED***);
    ***REMOVED***,
***REMOVED***);
