// @typescript-eslint/eslint-plugin/dist/rules/ban-ts-comment.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; ***REMOVED*** ***REMOVED***;
    ***REMOVED***
    Object.defineProperty(o, k2, desc);
***REMOVED***) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
***REMOVED***));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v ***REMOVED***);
***REMOVED***) : function(o, v) {
    o["default"] = v;
***REMOVED***);
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {***REMOVED***;
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.defaultMinimumDescriptionLength = void 0;
const utils_1 = require("@typescript-eslint/utils");
const util = __importStar(require("../util"));
exports.defaultMinimumDescriptionLength = 3;
exports.default = util.createRule({
    name: 'ban-ts-comment',
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallow `@ts-<directive>` comments or require descriptions after directives',
            recommended: 'error',
        ***REMOVED***,
        messages: {
            tsDirectiveComment: 'Do not use "@ts-{{directive***REMOVED******REMOVED***" because it alters compilation errors.',
            tsDirectiveCommentRequiresDescription: 'Include a description after the "@ts-{{directive***REMOVED******REMOVED***" directive to explain why the @ts-{{directive***REMOVED******REMOVED*** is necessary. The description must be {{minimumDescriptionLength***REMOVED******REMOVED*** characters or longer.',
            tsDirectiveCommentDescriptionNotMatchPattern: 'The description for the "@ts-{{directive***REMOVED******REMOVED***" directive must match the {{format***REMOVED******REMOVED*** format.',
        ***REMOVED***,
        schema: {
            $defs: {
                directiveConfigSchema: {
                    oneOf: [
                        {
                            type: 'boolean',
                            default: true,
                        ***REMOVED***,
                        {
                            enum: ['allow-with-description'],
                        ***REMOVED***,
                        {
                            type: 'object',
                            properties: {
                                descriptionFormat: { type: 'string' ***REMOVED***,
                            ***REMOVED***,
                        ***REMOVED***,
                    ],
                ***REMOVED***,
            ***REMOVED***,
            prefixItems: [
                {
                    properties: {
                        'ts-expect-error': {
                            $ref: '#/$defs/directiveConfigSchema',
                        ***REMOVED***,
                        'ts-ignore': { $ref: '#/$defs/directiveConfigSchema' ***REMOVED***,
                        'ts-nocheck': { $ref: '#/$defs/directiveConfigSchema' ***REMOVED***,
                        'ts-check': { $ref: '#/$defs/directiveConfigSchema' ***REMOVED***,
                        minimumDescriptionLength: {
                            type: 'number',
                            default: exports.defaultMinimumDescriptionLength,
                        ***REMOVED***,
                    ***REMOVED***,
                    additionalProperties: false,
                ***REMOVED***,
            ],
            type: 'array',
        ***REMOVED***,
    ***REMOVED***,
    defaultOptions: [
        {
            'ts-expect-error': 'allow-with-description',
            'ts-ignore': true,
            'ts-nocheck': true,
            'ts-check': false,
            minimumDescriptionLength: exports.defaultMinimumDescriptionLength,
        ***REMOVED***,
    ],
    create(context, [options]) {
        /*
          The regex used are taken from the ones used in the official TypeScript repo -
          https://github.com/microsoft/TypeScript/blob/408c760fae66080104bc85c449282c2d207dfe8e/src/compiler/scanner.ts#L288-L296
        */
        const commentDirectiveRegExSingleLine = /^\/*\s*@ts-(?<directive>expect-error|ignore|check|nocheck)(?<description>.*)/;
        const commentDirectiveRegExMultiLine = /^\s*(?:\/|\*)*\s*@ts-(?<directive>expect-error|ignore|check|nocheck)(?<description>.*)/;
        const sourceCode = context.getSourceCode();
        const descriptionFormats = new Map();
        for (const directive of [
            'ts-expect-error',
            'ts-ignore',
            'ts-nocheck',
            'ts-check',
        ]) {
            const option = options[directive];
            if (typeof option === 'object' && option.descriptionFormat) {
                descriptionFormats.set(directive, new RegExp(option.descriptionFormat));
            ***REMOVED***
        ***REMOVED***
        return {
            Program() {
                const comments = sourceCode.getAllComments();
                comments.forEach(comment => {
                    const regExp = comment.type === utils_1.AST_TOKEN_TYPES.Line
                        ? commentDirectiveRegExSingleLine
                        : commentDirectiveRegExMultiLine;
                    const match = regExp.exec(comment.value);
                    if (!match) {
                        return;
                    ***REMOVED***
                    const { directive, description ***REMOVED*** = match.groups;
                    const fullDirective = `ts-${directive***REMOVED***`;
                    const option = options[fullDirective];
                    if (option === true) {
                        context.report({
                            data: { directive ***REMOVED***,
                            node: comment,
                            messageId: 'tsDirectiveComment',
                        ***REMOVED***);
                    ***REMOVED***
                    if (option === 'allow-with-description' ||
                        (typeof option === 'object' && option.descriptionFormat)) {
                        const { minimumDescriptionLength = exports.defaultMinimumDescriptionLength, ***REMOVED*** = options;
                        const format = descriptionFormats.get(fullDirective);
                        if (util.getStringLength(description.trim()) <
                            minimumDescriptionLength) {
                            context.report({
                                data: { directive, minimumDescriptionLength ***REMOVED***,
                                node: comment,
                                messageId: 'tsDirectiveCommentRequiresDescription',
                            ***REMOVED***);
                        ***REMOVED***
                        else if (format && !format.test(description)) {
                            context.report({
                                data: { directive, format: format.source ***REMOVED***,
                                node: comment,
                                messageId: 'tsDirectiveCommentDescriptionNotMatchPattern',
                            ***REMOVED***);
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***,
        ***REMOVED***;
    ***REMOVED***,
***REMOVED***);
