// @typescript-eslint/eslint-plugin/dist/rules/consistent-type-imports.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; ***REMOVED*** ***REMOVED***;
    ***REMOVED***
    Object.defineProperty(o, k2, desc);
***REMOVED***) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
***REMOVED***));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v ***REMOVED***);
***REMOVED***) : function(o, v) {
    o["default"] = v;
***REMOVED***);
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {***REMOVED***;
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const utils_1 = require("@typescript-eslint/utils");
const util = __importStar(require("../util"));
exports.default = util.createRule({
    name: 'consistent-type-imports',
    meta: {
        type: 'suggestion',
        docs: {
            description: 'Enforce consistent usage of type imports',
            recommended: false,
        ***REMOVED***,
        messages: {
            typeOverValue: 'All imports in the declaration are only used as types. Use `import type`.',
            someImportsAreOnlyTypes: 'Imports {{typeImports***REMOVED******REMOVED*** are only used as types.',
            aImportIsOnlyTypes: 'Import {{typeImports***REMOVED******REMOVED*** is only used as types.',
            someImportsInDecoMeta: 'Type imports {{typeImports***REMOVED******REMOVED*** are used by decorator metadata.',
            aImportInDecoMeta: 'Type import {{typeImports***REMOVED******REMOVED*** is used by decorator metadata.',
            valueOverType: 'Use an `import` instead of an `import type`.',
            noImportTypeAnnotations: '`import()` type annotations are forbidden.',
        ***REMOVED***,
        schema: [
            {
                type: 'object',
                properties: {
                    prefer: {
                        enum: ['type-imports', 'no-type-imports'],
                    ***REMOVED***,
                    disallowTypeAnnotations: {
                        type: 'boolean',
                    ***REMOVED***,
                    fixStyle: {
                        enum: ['separate-type-imports', 'inline-type-imports'],
                    ***REMOVED***,
                ***REMOVED***,
                additionalProperties: false,
            ***REMOVED***,
        ],
        fixable: 'code',
    ***REMOVED***,
    defaultOptions: [
        {
            prefer: 'type-imports',
            disallowTypeAnnotations: true,
            fixStyle: 'separate-type-imports',
        ***REMOVED***,
    ],
    create(context, [option]) {
        var _a, _b;
        const prefer = (_a = option.prefer) !== null && _a !== void 0 ? _a : 'type-imports';
        const disallowTypeAnnotations = option.disallowTypeAnnotations !== false;
        const fixStyle = (_b = option.fixStyle) !== null && _b !== void 0 ? _b : 'separate-type-imports';
        const sourceCode = context.getSourceCode();
        const sourceImportsMap = {***REMOVED***;
        return Object.assign(Object.assign({***REMOVED***, (prefer === 'type-imports'
            ? {
                // prefer type imports
                ImportDeclaration(node) {
                    var _a;
                    const source = node.source.value;
                    // sourceImports is the object containing all the specifics for a particular import source, type or value
                    const sourceImports = (_a = sourceImportsMap[source]) !== null && _a !== void 0 ? _a : (sourceImportsMap[source] = {
                        source,
                        reportValueImports: [],
                        typeOnlyNamedImport: null,
                        valueOnlyNamedImport: null,
                        valueImport: null, // if only value imports
                    ***REMOVED***);
                    if (node.importKind === 'type') {
                        if (!sourceImports.typeOnlyNamedImport &&
                            node.specifiers.every(specifier => specifier.type === utils_1.AST_NODE_TYPES.ImportSpecifier)) {
                            // definitely import type { TypeX ***REMOVED***
                            sourceImports.typeOnlyNamedImport = node;
                        ***REMOVED***
                    ***REMOVED***
                    else {
                        if (!sourceImports.valueOnlyNamedImport &&
                            node.specifiers.every(specifier => specifier.type === utils_1.AST_NODE_TYPES.ImportSpecifier)) {
                            sourceImports.valueOnlyNamedImport = node;
                            sourceImports.valueImport = node;
                        ***REMOVED***
                        else if (!sourceImports.valueImport &&
                            node.specifiers.some(specifier => specifier.type === utils_1.AST_NODE_TYPES.ImportDefaultSpecifier)) {
                            sourceImports.valueImport = node;
                        ***REMOVED***
                    ***REMOVED***
                    const typeSpecifiers = [];
                    const inlineTypeSpecifiers = [];
                    const valueSpecifiers = [];
                    const unusedSpecifiers = [];
                    for (const specifier of node.specifiers) {
                        if (specifier.type === utils_1.AST_NODE_TYPES.ImportSpecifier &&
                            specifier.importKind === 'type') {
                            inlineTypeSpecifiers.push(specifier);
                            continue;
                        ***REMOVED***
                        const [variable] = context.getDeclaredVariables(specifier);
                        if (variable.references.length === 0) {
                            unusedSpecifiers.push(specifier);
                        ***REMOVED***
                        else {
                            const onlyHasTypeReferences = variable.references.every(ref => {
                                var _a, _b;
                                /**
                                 * keep origin import kind when export
                                 * export { Type ***REMOVED***
                                 * export default Type;
                                 */
                                if (((_a = ref.identifier.parent) === null || _a === void 0 ? void 0 : _a.type) ===
                                    utils_1.AST_NODE_TYPES.ExportSpecifier ||
                                    ((_b = ref.identifier.parent) === null || _b === void 0 ? void 0 : _b.type) ===
                                        utils_1.AST_NODE_TYPES.ExportDefaultDeclaration) {
                                    if (ref.isValueReference && ref.isTypeReference) {
                                        return node.importKind === 'type';
                                    ***REMOVED***
                                ***REMOVED***
                                if (ref.isValueReference) {
                                    let parent = ref.identifier.parent;
                                    let child = ref.identifier;
                                    while (parent) {
                                        switch (parent.type) {
                                            // CASE 1:
                                            // `type T = typeof foo` will create a value reference because "foo" must be a value type
                                            // however this value reference is safe to use with type-only imports
                                            case utils_1.AST_NODE_TYPES.TSTypeQuery:
                                                return true;
                                            case utils_1.AST_NODE_TYPES.TSQualifiedName:
                                                // TSTypeQuery must have a TSESTree.EntityName as its child, so we can filter here and break early
                                                if (parent.left !== child) {
                                                    return false;
                                                ***REMOVED***
                                                child = parent;
                                                parent = parent.parent;
                                                continue;
                                            // END CASE 1
                                            //////////////
                                            // CASE 2:
                                            // `type T = { [foo]: string ***REMOVED***` will create a value reference because "foo" must be a value type
                                            // however this value reference is safe to use with type-only imports.
                                            // Also this is represented as a non-type AST - hence it uses MemberExpression
                                            case utils_1.AST_NODE_TYPES.TSPropertySignature:
                                                return parent.key === child;
                                            case utils_1.AST_NODE_TYPES.MemberExpression:
                                                if (parent.object !== child) {
                                                    return false;
                                                ***REMOVED***
                                                child = parent;
                                                parent = parent.parent;
                                                continue;
                                            // END CASE 2
                                            default:
                                                return false;
                                        ***REMOVED***
                                    ***REMOVED***
                                ***REMOVED***
                                return ref.isTypeReference;
                            ***REMOVED***);
                            if (onlyHasTypeReferences) {
                                typeSpecifiers.push(specifier);
                            ***REMOVED***
                            else {
                                valueSpecifiers.push(specifier);
                            ***REMOVED***
                        ***REMOVED***
                    ***REMOVED***
                    if ((node.importKind === 'value' && typeSpecifiers.length) ||
                        (node.importKind === 'type' && valueSpecifiers.length)) {
                        sourceImports.reportValueImports.push({
                            node,
                            typeSpecifiers,
                            valueSpecifiers,
                            unusedSpecifiers,
                            inlineTypeSpecifiers,
                        ***REMOVED***);
                    ***REMOVED***
                ***REMOVED***,
                'Program:exit'() {
                    for (const sourceImports of Object.values(sourceImportsMap)) {
                        if (sourceImports.reportValueImports.length === 0) {
                            // nothing to fix. value specifiers and type specifiers are correctly written
                            continue;
                        ***REMOVED***
                        for (const report of sourceImports.reportValueImports) {
                            if (report.valueSpecifiers.length === 0 &&
                                report.unusedSpecifiers.length === 0 &&
                                report.node.importKind !== 'type') {
                                context.report({
                                    node: report.node,
                                    messageId: 'typeOverValue',
                                    *fix(fixer) {
                                        yield* fixToTypeImportDeclaration(fixer, report, sourceImports);
                                    ***REMOVED***,
                                ***REMOVED***);
                            ***REMOVED***
                            else {
                                const isTypeImport = report.node.importKind === 'type';
                                // we have a mixed type/value import or just value imports, so we need to split them out into multiple imports if separate-type-imports is configured
                                const importNames = (isTypeImport
                                    ? report.valueSpecifiers // import type { A ***REMOVED*** from 'roo'; // WHERE A is used in value position
                                    : report.typeSpecifiers) // import { A, B ***REMOVED*** from 'roo'; // WHERE A is used in type position and B is in value position
                                    .map(specifier => `"${specifier.local.name***REMOVED***"`);
                                const message = (() => {
                                    const typeImports = util.formatWordList(importNames);
                                    if (importNames.length === 1) {
                                        if (isTypeImport) {
                                            return {
                                                messageId: 'aImportInDecoMeta',
                                                data: { typeImports ***REMOVED***,
                                            ***REMOVED***;
                                        ***REMOVED***
                                        else {
                                            return {
                                                messageId: 'aImportIsOnlyTypes',
                                                data: { typeImports ***REMOVED***,
                                            ***REMOVED***;
                                        ***REMOVED***
                                    ***REMOVED***
                                    else {
                                        if (isTypeImport) {
                                            return {
                                                messageId: 'someImportsInDecoMeta',
                                                data: { typeImports ***REMOVED***, // typeImports are all the value specifiers that are in the type position
                                            ***REMOVED***;
                                        ***REMOVED***
                                        else {
                                            return {
                                                messageId: 'someImportsAreOnlyTypes',
                                                data: { typeImports ***REMOVED***, // typeImports are all the type specifiers in the value position
                                            ***REMOVED***;
                                        ***REMOVED***
                                    ***REMOVED***
                                ***REMOVED***)();
                                context.report(Object.assign(Object.assign({ node: report.node ***REMOVED***, message), { *fix(fixer) {
                                        if (isTypeImport) {
                                            // take all the valueSpecifiers and put them on a new line
                                            yield* fixToValueImportDeclaration(fixer, report, sourceImports);
                                        ***REMOVED***
                                        else {
                                            // take all the typeSpecifiers and put them on a new line
                                            yield* fixToTypeImportDeclaration(fixer, report, sourceImports);
                                        ***REMOVED***
                                    ***REMOVED*** ***REMOVED***));
                            ***REMOVED***
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***,
            ***REMOVED***
            : {
                // prefer no type imports
                'ImportDeclaration[importKind = "type"]'(node) {
                    context.report({
                        node,
                        messageId: 'valueOverType',
                        fix(fixer) {
                            return fixRemoveTypeSpecifierFromImportDeclaration(fixer, node);
                        ***REMOVED***,
                    ***REMOVED***);
                ***REMOVED***,
                'ImportSpecifier[importKind = "type"]'(node) {
                    context.report({
                        node,
                        messageId: 'valueOverType',
                        fix(fixer) {
                            return fixRemoveTypeSpecifierFromImportSpecifier(fixer, node);
                        ***REMOVED***,
                    ***REMOVED***);
                ***REMOVED***,
            ***REMOVED***)), (disallowTypeAnnotations
            ? {
                // disallow `import()` type
                TSImportType(node) {
                    context.report({
                        node,
                        messageId: 'noImportTypeAnnotations',
                    ***REMOVED***);
                ***REMOVED***,
            ***REMOVED***
            : {***REMOVED***));
        function classifySpecifier(node) {
            var _a;
            const defaultSpecifier = node.specifiers[0].type === utils_1.AST_NODE_TYPES.ImportDefaultSpecifier
                ? node.specifiers[0]
                : null;
            const namespaceSpecifier = (_a = node.specifiers.find((specifier) => specifier.type === utils_1.AST_NODE_TYPES.ImportNamespaceSpecifier)) !== null && _a !== void 0 ? _a : null;
            const namedSpecifiers = node.specifiers.filter((specifier) => specifier.type === utils_1.AST_NODE_TYPES.ImportSpecifier);
            return {
                defaultSpecifier,
                namespaceSpecifier,
                namedSpecifiers,
            ***REMOVED***;
        ***REMOVED***
        /**
         * Returns information for fixing named specifiers, type or value
         */
        function getFixesNamedSpecifiers(fixer, node, subsetNamedSpecifiers, allNamedSpecifiers) {
            if (allNamedSpecifiers.length === 0) {
                return {
                    typeNamedSpecifiersText: '',
                    removeTypeNamedSpecifiers: [],
                ***REMOVED***;
            ***REMOVED***
            const typeNamedSpecifiersTexts = [];
            const removeTypeNamedSpecifiers = [];
            if (subsetNamedSpecifiers.length === allNamedSpecifiers.length) {
                // import Foo, {Type1, Type2***REMOVED*** from 'foo'
                // import DefType, {Type1, Type2***REMOVED*** from 'foo'
                const openingBraceToken = util.nullThrows(sourceCode.getTokenBefore(subsetNamedSpecifiers[0], util.isOpeningBraceToken), util.NullThrowsReasons.MissingToken('{', node.type));
                const commaToken = util.nullThrows(sourceCode.getTokenBefore(openingBraceToken, util.isCommaToken), util.NullThrowsReasons.MissingToken(',', node.type));
                const closingBraceToken = util.nullThrows(sourceCode.getFirstTokenBetween(openingBraceToken, node.source, util.isClosingBraceToken), util.NullThrowsReasons.MissingToken('***REMOVED***', node.type));
                // import DefType, {...***REMOVED*** from 'foo'
                //               ^^^^^^^ remove
                removeTypeNamedSpecifiers.push(fixer.removeRange([commaToken.range[0], closingBraceToken.range[1]]));
                typeNamedSpecifiersTexts.push(sourceCode.text.slice(openingBraceToken.range[1], closingBraceToken.range[0]));
            ***REMOVED***
            else {
                const namedSpecifierGroups = [];
                let group = [];
                for (const namedSpecifier of allNamedSpecifiers) {
                    if (subsetNamedSpecifiers.includes(namedSpecifier)) {
                        group.push(namedSpecifier);
                    ***REMOVED***
                    else if (group.length) {
                        namedSpecifierGroups.push(group);
                        group = [];
                    ***REMOVED***
                ***REMOVED***
                if (group.length) {
                    namedSpecifierGroups.push(group);
                ***REMOVED***
                for (const namedSpecifiers of namedSpecifierGroups) {
                    const { removeRange, textRange ***REMOVED*** = getNamedSpecifierRanges(namedSpecifiers, allNamedSpecifiers);
                    removeTypeNamedSpecifiers.push(fixer.removeRange(removeRange));
                    typeNamedSpecifiersTexts.push(sourceCode.text.slice(...textRange));
                ***REMOVED***
            ***REMOVED***
            return {
                typeNamedSpecifiersText: typeNamedSpecifiersTexts.join(','),
                removeTypeNamedSpecifiers,
            ***REMOVED***;
        ***REMOVED***
        /**
         * Returns ranges for fixing named specifier.
         */
        function getNamedSpecifierRanges(namedSpecifierGroup, allNamedSpecifiers) {
            const first = namedSpecifierGroup[0];
            const last = namedSpecifierGroup[namedSpecifierGroup.length - 1];
            const removeRange = [first.range[0], last.range[1]];
            const textRange = [...removeRange];
            const before = sourceCode.getTokenBefore(first);
            textRange[0] = before.range[1];
            if (util.isCommaToken(before)) {
                removeRange[0] = before.range[0];
            ***REMOVED***
            else {
                removeRange[0] = before.range[1];
            ***REMOVED***
            const isFirst = allNamedSpecifiers[0] === first;
            const isLast = allNamedSpecifiers[allNamedSpecifiers.length - 1] === last;
            const after = sourceCode.getTokenAfter(last);
            textRange[1] = after.range[0];
            if (isFirst || isLast) {
                if (util.isCommaToken(after)) {
                    removeRange[1] = after.range[1];
                ***REMOVED***
            ***REMOVED***
            return {
                textRange,
                removeRange,
            ***REMOVED***;
        ***REMOVED***
        /**
         * insert specifiers to named import node.
         * e.g.
         * import type { Already, Type1, Type2 ***REMOVED*** from 'foo'
         *                        ^^^^^^^^^^^^^ insert
         */
        function fixInsertNamedSpecifiersInNamedSpecifierList(fixer, target, insertText) {
            const closingBraceToken = util.nullThrows(sourceCode.getFirstTokenBetween(sourceCode.getFirstToken(target), target.source, util.isClosingBraceToken), util.NullThrowsReasons.MissingToken('***REMOVED***', target.type));
            const before = sourceCode.getTokenBefore(closingBraceToken);
            if (!util.isCommaToken(before) && !util.isOpeningBraceToken(before)) {
                insertText = `,${insertText***REMOVED***`;
            ***REMOVED***
            return fixer.insertTextBefore(closingBraceToken, insertText);
        ***REMOVED***
        /**
         * insert type keyword to named import node.
         * e.g.
         * import ADefault, { Already, type Type1, type Type2 ***REMOVED*** from 'foo'
         *                             ^^^^ insert
         */
        function* fixInsertTypeKeywordInNamedSpecifierList(fixer, typeSpecifiers) {
            for (const spec of typeSpecifiers) {
                const insertText = sourceCode.text.slice(...spec.range);
                yield fixer.replaceTextRange(spec.range, `type ${insertText***REMOVED***`);
            ***REMOVED***
        ***REMOVED***
        function* fixInlineTypeImportDeclaration(fixer, report, sourceImports) {
            const { node ***REMOVED*** = report;
            // For a value import, will only add an inline type to named specifiers
            const { namedSpecifiers ***REMOVED*** = classifySpecifier(node);
            const typeNamedSpecifiers = namedSpecifiers.filter(specifier => report.typeSpecifiers.includes(specifier));
            if (sourceImports.valueImport) {
                // add import named type specifiers to its value import
                // import ValueA, { type A ***REMOVED***
                //                  ^^^^ insert
                const { namedSpecifiers: valueImportNamedSpecifiers ***REMOVED*** = classifySpecifier(sourceImports.valueImport);
                if (sourceImports.valueOnlyNamedImport ||
                    valueImportNamedSpecifiers.length) {
                    yield* fixInsertTypeKeywordInNamedSpecifierList(fixer, typeNamedSpecifiers);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        function* fixToTypeImportDeclaration(fixer, report, sourceImports) {
            const { node ***REMOVED*** = report;
            const { defaultSpecifier, namespaceSpecifier, namedSpecifiers ***REMOVED*** = classifySpecifier(node);
            if (namespaceSpecifier && !defaultSpecifier) {
                // import * as types from 'foo'
                yield* fixInsertTypeSpecifierForImportDeclaration(fixer, node, false);
                return;
            ***REMOVED***
            else if (defaultSpecifier) {
                if (report.typeSpecifiers.includes(defaultSpecifier) &&
                    namedSpecifiers.length === 0 &&
                    !namespaceSpecifier) {
                    // import Type from 'foo'
                    yield* fixInsertTypeSpecifierForImportDeclaration(fixer, node, true);
                    return;
                ***REMOVED***
                else if (fixStyle === 'inline-type-imports' &&
                    !report.typeSpecifiers.includes(defaultSpecifier) &&
                    namedSpecifiers.length > 0 &&
                    !namespaceSpecifier) {
                    // if there is a default specifier but it isn't a type specifier, then just add the inline type modifier to the named specifiers
                    // import AValue, {BValue, Type1, Type2***REMOVED*** from 'foo'
                    yield* fixInlineTypeImportDeclaration(fixer, report, sourceImports);
                    return;
                ***REMOVED***
            ***REMOVED***
            else if (!namespaceSpecifier) {
                if (fixStyle === 'inline-type-imports' &&
                    namedSpecifiers.some(specifier => report.typeSpecifiers.includes(specifier))) {
                    // import {AValue, Type1, Type2***REMOVED*** from 'foo'
                    yield* fixInlineTypeImportDeclaration(fixer, report, sourceImports);
                    return;
                ***REMOVED***
                else if (namedSpecifiers.every(specifier => report.typeSpecifiers.includes(specifier))) {
                    // import {Type1, Type2***REMOVED*** from 'foo'
                    yield* fixInsertTypeSpecifierForImportDeclaration(fixer, node, false);
                    return;
                ***REMOVED***
            ***REMOVED***
            const typeNamedSpecifiers = namedSpecifiers.filter(specifier => report.typeSpecifiers.includes(specifier));
            const fixesNamedSpecifiers = getFixesNamedSpecifiers(fixer, node, typeNamedSpecifiers, namedSpecifiers);
            const afterFixes = [];
            if (typeNamedSpecifiers.length) {
                if (sourceImports.typeOnlyNamedImport) {
                    const insertTypeNamedSpecifiers = fixInsertNamedSpecifiersInNamedSpecifierList(fixer, sourceImports.typeOnlyNamedImport, fixesNamedSpecifiers.typeNamedSpecifiersText);
                    if (sourceImports.typeOnlyNamedImport.range[1] <= node.range[0]) {
                        yield insertTypeNamedSpecifiers;
                    ***REMOVED***
                    else {
                        afterFixes.push(insertTypeNamedSpecifiers);
                    ***REMOVED***
                ***REMOVED***
                else {
                    // The import is both default and named.  Insert named on new line because can't mix default type import and named type imports
                    if (fixStyle === 'inline-type-imports') {
                        yield fixer.insertTextBefore(node, `import {${typeNamedSpecifiers
                            .map(spec => {
                            const insertText = sourceCode.text.slice(...spec.range);
                            return `type ${insertText***REMOVED***`;
                        ***REMOVED***)
                            .join(', ')***REMOVED******REMOVED*** from ${sourceCode.getText(node.source)***REMOVED***;\n`);
                    ***REMOVED***
                    else {
                        yield fixer.insertTextBefore(node, `import type {${fixesNamedSpecifiers.typeNamedSpecifiersText***REMOVED******REMOVED*** from ${sourceCode.getText(node.source)***REMOVED***;\n`);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
            const fixesRemoveTypeNamespaceSpecifier = [];
            if (namespaceSpecifier &&
                report.typeSpecifiers.includes(namespaceSpecifier)) {
                // import Foo, * as Type from 'foo'
                // import DefType, * as Type from 'foo'
                // import DefType, * as Type from 'foo'
                const commaToken = util.nullThrows(sourceCode.getTokenBefore(namespaceSpecifier, util.isCommaToken), util.NullThrowsReasons.MissingToken(',', node.type));
                // import Def, * as Ns from 'foo'
                //           ^^^^^^^^^ remove
                fixesRemoveTypeNamespaceSpecifier.push(fixer.removeRange([commaToken.range[0], namespaceSpecifier.range[1]]));
                // import type * as Ns from 'foo'
                // ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ insert
                yield fixer.insertTextBefore(node, `import type ${sourceCode.getText(namespaceSpecifier)***REMOVED*** from ${sourceCode.getText(node.source)***REMOVED***;\n`);
            ***REMOVED***
            if (defaultSpecifier &&
                report.typeSpecifiers.includes(defaultSpecifier)) {
                if (report.typeSpecifiers.length === node.specifiers.length) {
                    const importToken = util.nullThrows(sourceCode.getFirstToken(node, util.isImportKeyword), util.NullThrowsReasons.MissingToken('import', node.type));
                    // import type Type from 'foo'
                    //        ^^^^ insert
                    yield fixer.insertTextAfter(importToken, ' type');
                ***REMOVED***
                else {
                    const commaToken = util.nullThrows(sourceCode.getTokenAfter(defaultSpecifier, util.isCommaToken), util.NullThrowsReasons.MissingToken(',', defaultSpecifier.type));
                    // import Type , {...***REMOVED*** from 'foo'
                    //        ^^^^^ pick
                    const defaultText = sourceCode.text
                        .slice(defaultSpecifier.range[0], commaToken.range[0])
                        .trim();
                    yield fixer.insertTextBefore(node, `import type ${defaultText***REMOVED*** from ${sourceCode.getText(node.source)***REMOVED***;\n`);
                    const afterToken = util.nullThrows(sourceCode.getTokenAfter(commaToken, { includeComments: true ***REMOVED***), util.NullThrowsReasons.MissingToken('any token', node.type));
                    // import Type , {...***REMOVED*** from 'foo'
                    //        ^^^^^^^ remove
                    yield fixer.removeRange([
                        defaultSpecifier.range[0],
                        afterToken.range[0],
                    ]);
                ***REMOVED***
            ***REMOVED***
            yield* fixesNamedSpecifiers.removeTypeNamedSpecifiers;
            yield* fixesRemoveTypeNamespaceSpecifier;
            yield* afterFixes;
        ***REMOVED***
        function* fixInsertTypeSpecifierForImportDeclaration(fixer, node, isDefaultImport) {
            // import type Foo from 'foo'
            //       ^^^^^ insert
            const importToken = util.nullThrows(sourceCode.getFirstToken(node, util.isImportKeyword), util.NullThrowsReasons.MissingToken('import', node.type));
            yield fixer.insertTextAfter(importToken, ' type');
            if (isDefaultImport) {
                // Has default import
                const openingBraceToken = sourceCode.getFirstTokenBetween(importToken, node.source, util.isOpeningBraceToken);
                if (openingBraceToken) {
                    // Only braces. e.g. import Foo, {***REMOVED*** from 'foo'
                    const commaToken = util.nullThrows(sourceCode.getTokenBefore(openingBraceToken, util.isCommaToken), util.NullThrowsReasons.MissingToken(',', node.type));
                    const closingBraceToken = util.nullThrows(sourceCode.getFirstTokenBetween(openingBraceToken, node.source, util.isClosingBraceToken), util.NullThrowsReasons.MissingToken('***REMOVED***', node.type));
                    // import type Foo, {***REMOVED*** from 'foo'
                    //                  ^^ remove
                    yield fixer.removeRange([
                        commaToken.range[0],
                        closingBraceToken.range[1],
                    ]);
                    const specifiersText = sourceCode.text.slice(commaToken.range[1], closingBraceToken.range[1]);
                    if (node.specifiers.length > 1) {
                        yield fixer.insertTextAfter(node, `\nimport type${specifiersText***REMOVED*** from ${sourceCode.getText(node.source)***REMOVED***;`);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
            // make sure we don't do anything like `import type {type T***REMOVED*** from 'foo';`
            for (const specifier of node.specifiers) {
                if (specifier.type === utils_1.AST_NODE_TYPES.ImportSpecifier &&
                    specifier.importKind === 'type') {
                    yield* fixRemoveTypeSpecifierFromImportSpecifier(fixer, specifier);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        function* fixToValueImportDeclaration(fixer, report, sourceImports) {
            const { node ***REMOVED*** = report;
            const { defaultSpecifier, namespaceSpecifier, namedSpecifiers ***REMOVED*** = classifySpecifier(node);
            if (namespaceSpecifier) {
                // import type * as types from 'foo'
                yield* fixRemoveTypeSpecifierFromImportDeclaration(fixer, node);
                return;
            ***REMOVED***
            else if (defaultSpecifier) {
                if (report.valueSpecifiers.includes(defaultSpecifier) &&
                    namedSpecifiers.length === 0) {
                    // import type Type from 'foo'
                    yield* fixRemoveTypeSpecifierFromImportDeclaration(fixer, node);
                    return;
                ***REMOVED***
            ***REMOVED***
            else {
                if (namedSpecifiers.every(specifier => report.valueSpecifiers.includes(specifier))) {
                    // import type {Type1, Type2***REMOVED*** from 'foo'
                    yield* fixRemoveTypeSpecifierFromImportDeclaration(fixer, node);
                    return;
                ***REMOVED***
            ***REMOVED***
            // we have some valueSpecifiers intermixed in types that need to be put on their own line
            // import type { Type1, A ***REMOVED*** from 'foo'
            // import type { A ***REMOVED*** from 'foo'
            const valueNamedSpecifiers = namedSpecifiers.filter(specifier => report.valueSpecifiers.includes(specifier));
            const fixesNamedSpecifiers = getFixesNamedSpecifiers(fixer, node, valueNamedSpecifiers, namedSpecifiers);
            const afterFixes = [];
            if (valueNamedSpecifiers.length) {
                if (sourceImports.valueOnlyNamedImport) {
                    const insertTypeNamedSpecifiers = fixInsertNamedSpecifiersInNamedSpecifierList(fixer, sourceImports.valueOnlyNamedImport, fixesNamedSpecifiers.typeNamedSpecifiersText);
                    if (sourceImports.valueOnlyNamedImport.range[1] <= node.range[0]) {
                        yield insertTypeNamedSpecifiers;
                    ***REMOVED***
                    else {
                        afterFixes.push(insertTypeNamedSpecifiers);
                    ***REMOVED***
                ***REMOVED***
                else {
                    // some are types.
                    // Add new value import and later remove those value specifiers from import type
                    yield fixer.insertTextBefore(node, `import {${fixesNamedSpecifiers.typeNamedSpecifiersText***REMOVED******REMOVED*** from ${sourceCode.getText(node.source)***REMOVED***;\n`);
                ***REMOVED***
            ***REMOVED***
            yield* fixesNamedSpecifiers.removeTypeNamedSpecifiers;
            yield* afterFixes;
        ***REMOVED***
        function* fixRemoveTypeSpecifierFromImportDeclaration(fixer, node) {
            var _a, _b;
            // import type Foo from 'foo'
            //        ^^^^ remove
            const importToken = util.nullThrows(sourceCode.getFirstToken(node, util.isImportKeyword), util.NullThrowsReasons.MissingToken('import', node.type));
            const typeToken = util.nullThrows(sourceCode.getFirstTokenBetween(importToken, (_b = (_a = node.specifiers[0]) === null || _a === void 0 ? void 0 : _a.local) !== null && _b !== void 0 ? _b : node.source, util.isTypeKeyword), util.NullThrowsReasons.MissingToken('type', node.type));
            const afterToken = util.nullThrows(sourceCode.getTokenAfter(typeToken, { includeComments: true ***REMOVED***), util.NullThrowsReasons.MissingToken('any token', node.type));
            yield fixer.removeRange([typeToken.range[0], afterToken.range[0]]);
        ***REMOVED***
        function* fixRemoveTypeSpecifierFromImportSpecifier(fixer, node) {
            // import { type Foo ***REMOVED*** from 'foo'
            //          ^^^^ remove
            const typeToken = util.nullThrows(sourceCode.getFirstToken(node, util.isTypeKeyword), util.NullThrowsReasons.MissingToken('type', node.type));
            const afterToken = util.nullThrows(sourceCode.getTokenAfter(typeToken, { includeComments: true ***REMOVED***), util.NullThrowsReasons.MissingToken('any token', node.type));
            yield fixer.removeRange([typeToken.range[0], afterToken.range[0]]);
        ***REMOVED***
    ***REMOVED***,
***REMOVED***);
