// @typescript-eslint/eslint-plugin/dist/rules/explicit-member-accessibility.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; ***REMOVED*** ***REMOVED***;
    ***REMOVED***
    Object.defineProperty(o, k2, desc);
***REMOVED***) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
***REMOVED***));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v ***REMOVED***);
***REMOVED***) : function(o, v) {
    o["default"] = v;
***REMOVED***);
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {***REMOVED***;
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const utils_1 = require("@typescript-eslint/utils");
const util = __importStar(require("../util"));
const accessibilityLevel = {
    oneOf: [
        {
            const: 'explicit',
            description: 'Always require an accessor.',
        ***REMOVED***,
        {
            const: 'no-public',
            description: 'Require an accessor except when public.',
        ***REMOVED***,
        {
            const: 'off',
            description: 'Never check whether there is an accessor.',
        ***REMOVED***,
    ],
***REMOVED***;
exports.default = util.createRule({
    name: 'explicit-member-accessibility',
    meta: {
        hasSuggestions: true,
        type: 'problem',
        docs: {
            description: 'Require explicit accessibility modifiers on class properties and methods',
            // too opinionated to be recommended
            recommended: false,
        ***REMOVED***,
        fixable: 'code',
        messages: {
            missingAccessibility: 'Missing accessibility modifier on {{type***REMOVED******REMOVED*** {{name***REMOVED******REMOVED***.',
            unwantedPublicAccessibility: 'Public accessibility modifier on {{type***REMOVED******REMOVED*** {{name***REMOVED******REMOVED***.',
            addExplicitAccessibility: "Add '{{ type ***REMOVED******REMOVED***' accessibility modifier",
        ***REMOVED***,
        schema: {
            $defs: {
                accessibilityLevel,
            ***REMOVED***,
            prefixItems: [
                {
                    type: 'object',
                    properties: {
                        accessibility: { $ref: '#/$defs/accessibilityLevel' ***REMOVED***,
                        overrides: {
                            type: 'object',
                            properties: {
                                accessors: { $ref: '#/$defs/accessibilityLevel' ***REMOVED***,
                                constructors: { $ref: '#/$defs/accessibilityLevel' ***REMOVED***,
                                methods: { $ref: '#/$defs/accessibilityLevel' ***REMOVED***,
                                properties: { $ref: '#/$defs/accessibilityLevel' ***REMOVED***,
                                parameterProperties: {
                                    $ref: '#/$defs/accessibilityLevel',
                                ***REMOVED***,
                            ***REMOVED***,
                            additionalProperties: false,
                        ***REMOVED***,
                        ignoredMethodNames: {
                            type: 'array',
                            items: {
                                type: 'string',
                            ***REMOVED***,
                        ***REMOVED***,
                    ***REMOVED***,
                    additionalProperties: false,
                ***REMOVED***,
            ],
            type: 'array',
        ***REMOVED***,
    ***REMOVED***,
    defaultOptions: [{ accessibility: 'explicit' ***REMOVED***],
    create(context, [option]) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        const sourceCode = context.getSourceCode();
        const baseCheck = (_a = option.accessibility) !== null && _a !== void 0 ? _a : 'explicit';
        const overrides = (_b = option.overrides) !== null && _b !== void 0 ? _b : {***REMOVED***;
        const ctorCheck = (_c = overrides.constructors) !== null && _c !== void 0 ? _c : baseCheck;
        const accessorCheck = (_d = overrides.accessors) !== null && _d !== void 0 ? _d : baseCheck;
        const methodCheck = (_e = overrides.methods) !== null && _e !== void 0 ? _e : baseCheck;
        const propCheck = (_f = overrides.properties) !== null && _f !== void 0 ? _f : baseCheck;
        const paramPropCheck = (_g = overrides.parameterProperties) !== null && _g !== void 0 ? _g : baseCheck;
        const ignoredMethodNames = new Set((_h = option.ignoredMethodNames) !== null && _h !== void 0 ? _h : []);
        /**
         * Checks if a method declaration has an accessibility modifier.
         * @param methodDefinition The node representing a MethodDefinition.
         */
        function checkMethodAccessibilityModifier(methodDefinition) {
            if (methodDefinition.key.type === utils_1.AST_NODE_TYPES.PrivateIdentifier) {
                return;
            ***REMOVED***
            let nodeType = 'method definition';
            let check = baseCheck;
            switch (methodDefinition.kind) {
                case 'method':
                    check = methodCheck;
                    break;
                case 'constructor':
                    check = ctorCheck;
                    break;
                case 'get':
                case 'set':
                    check = accessorCheck;
                    nodeType = `${methodDefinition.kind***REMOVED*** property accessor`;
                    break;
            ***REMOVED***
            const { name: methodName ***REMOVED*** = util.getNameFromMember(methodDefinition, sourceCode);
            if (check === 'off' || ignoredMethodNames.has(methodName)) {
                return;
            ***REMOVED***
            if (check === 'no-public' &&
                methodDefinition.accessibility === 'public') {
                context.report({
                    node: methodDefinition,
                    messageId: 'unwantedPublicAccessibility',
                    data: {
                        type: nodeType,
                        name: methodName,
                    ***REMOVED***,
                    fix: getUnwantedPublicAccessibilityFixer(methodDefinition),
                ***REMOVED***);
            ***REMOVED***
            else if (check === 'explicit' && !methodDefinition.accessibility) {
                context.report({
                    node: methodDefinition,
                    messageId: 'missingAccessibility',
                    data: {
                        type: nodeType,
                        name: methodName,
                    ***REMOVED***,
                    suggest: getMissingAccessibilitySuggestions(methodDefinition),
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***
        /**
         * Creates a fixer that removes a "public" keyword with following spaces
         */
        function getUnwantedPublicAccessibilityFixer(node) {
            return function (fixer) {
                const tokens = sourceCode.getTokens(node);
                let rangeToRemove;
                for (let i = 0; i < tokens.length; i++) {
                    const token = tokens[i];
                    if (token.type === utils_1.AST_TOKEN_TYPES.Keyword &&
                        token.value === 'public') {
                        const commensAfterPublicKeyword = sourceCode.getCommentsAfter(token);
                        if (commensAfterPublicKeyword.length) {
                            // public /* Hi there! */ static foo()
                            // ^^^^^^^
                            rangeToRemove = [
                                token.range[0],
                                commensAfterPublicKeyword[0].range[0],
                            ];
                            break;
                        ***REMOVED***
                        else {
                            // public static foo()
                            // ^^^^^^^
                            rangeToRemove = [token.range[0], tokens[i + 1].range[0]];
                            break;
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
                return fixer.removeRange(rangeToRemove);
            ***REMOVED***;
        ***REMOVED***
        /**
         * Creates a fixer that adds a "public" keyword with following spaces
         */
        function getMissingAccessibilitySuggestions(node) {
            function fix(accessibility, fixer) {
                var _a;
                if ((_a = node === null || node === void 0 ? void 0 : node.decorators) === null || _a === void 0 ? void 0 : _a.length) {
                    const lastDecorator = node.decorators[node.decorators.length - 1];
                    const nextToken = sourceCode.getTokenAfter(lastDecorator);
                    return fixer.insertTextBefore(nextToken, `${accessibility***REMOVED*** `);
                ***REMOVED***
                return fixer.insertTextBefore(node, `${accessibility***REMOVED*** `);
            ***REMOVED***
            return [
                {
                    messageId: 'addExplicitAccessibility',
                    data: { type: 'public' ***REMOVED***,
                    fix: fixer => fix('public', fixer),
                ***REMOVED***,
                {
                    messageId: 'addExplicitAccessibility',
                    data: { type: 'private' ***REMOVED***,
                    fix: fixer => fix('private', fixer),
                ***REMOVED***,
                {
                    messageId: 'addExplicitAccessibility',
                    data: { type: 'protected' ***REMOVED***,
                    fix: fixer => fix('protected', fixer),
                ***REMOVED***,
            ];
        ***REMOVED***
        /**
         * Checks if property has an accessibility modifier.
         * @param propertyDefinition The node representing a PropertyDefinition.
         */
        function checkPropertyAccessibilityModifier(propertyDefinition) {
            if (propertyDefinition.key.type === utils_1.AST_NODE_TYPES.PrivateIdentifier) {
                return;
            ***REMOVED***
            const nodeType = 'class property';
            const { name: propertyName ***REMOVED*** = util.getNameFromMember(propertyDefinition, sourceCode);
            if (propCheck === 'no-public' &&
                propertyDefinition.accessibility === 'public') {
                context.report({
                    node: propertyDefinition,
                    messageId: 'unwantedPublicAccessibility',
                    data: {
                        type: nodeType,
                        name: propertyName,
                    ***REMOVED***,
                    fix: getUnwantedPublicAccessibilityFixer(propertyDefinition),
                ***REMOVED***);
            ***REMOVED***
            else if (propCheck === 'explicit' &&
                !propertyDefinition.accessibility) {
                context.report({
                    node: propertyDefinition,
                    messageId: 'missingAccessibility',
                    data: {
                        type: nodeType,
                        name: propertyName,
                    ***REMOVED***,
                    suggest: getMissingAccessibilitySuggestions(propertyDefinition),
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***
        /**
         * Checks that the parameter property has the desired accessibility modifiers set.
         * @param node The node representing a Parameter Property
         */
        function checkParameterPropertyAccessibilityModifier(node) {
            const nodeType = 'parameter property';
            // HAS to be an identifier or assignment or TSC will throw
            if (node.parameter.type !== utils_1.AST_NODE_TYPES.Identifier &&
                node.parameter.type !== utils_1.AST_NODE_TYPES.AssignmentPattern) {
                return;
            ***REMOVED***
            const nodeName = node.parameter.type === utils_1.AST_NODE_TYPES.Identifier
                ? node.parameter.name
                : // has to be an Identifier or TSC will throw an error
                    node.parameter.left.name;
            switch (paramPropCheck) {
                case 'explicit': {
                    if (!node.accessibility) {
                        context.report({
                            node,
                            messageId: 'missingAccessibility',
                            data: {
                                type: nodeType,
                                name: nodeName,
                            ***REMOVED***,
                            suggest: getMissingAccessibilitySuggestions(node),
                        ***REMOVED***);
                    ***REMOVED***
                    break;
                ***REMOVED***
                case 'no-public': {
                    if (node.accessibility === 'public' && node.readonly) {
                        context.report({
                            node,
                            messageId: 'unwantedPublicAccessibility',
                            data: {
                                type: nodeType,
                                name: nodeName,
                            ***REMOVED***,
                            fix: getUnwantedPublicAccessibilityFixer(node),
                        ***REMOVED***);
                    ***REMOVED***
                    break;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        return {
            'MethodDefinition, TSAbstractMethodDefinition': checkMethodAccessibilityModifier,
            'PropertyDefinition, TSAbstractPropertyDefinition': checkPropertyAccessibilityModifier,
            TSParameterProperty: checkParameterPropertyAccessibilityModifier,
        ***REMOVED***;
    ***REMOVED***,
***REMOVED***);
