// @typescript-eslint/eslint-plugin/dist/rules/init-declarations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const utils_1 = require("@typescript-eslint/utils");
const util_1 = require("../util");
const getESLintCoreRule_1 = require("../util/getESLintCoreRule");
const baseRule = (0, getESLintCoreRule_1.getESLintCoreRule)('init-declarations');
exports.default = (0, util_1.createRule)({
    name: 'init-declarations',
    meta: {
        type: 'suggestion',
        docs: {
            description: 'Require or disallow initialization in variable declarations',
            recommended: false,
            extendsBaseRule: true,
        ***REMOVED***,
        hasSuggestions: baseRule.meta.hasSuggestions,
        schema: baseRule.meta.schema,
        messages: baseRule.meta.messages,
    ***REMOVED***,
    defaultOptions: ['always'],
    create(context, [mode]) {
        const rules = baseRule.create(context);
        return {
            'VariableDeclaration:exit'(node) {
                if (mode === 'always') {
                    if (node.declare) {
                        return;
                    ***REMOVED***
                    if (isAncestorNamespaceDeclared(node)) {
                        return;
                    ***REMOVED***
                ***REMOVED***
                rules['VariableDeclaration:exit'](node);
            ***REMOVED***,
        ***REMOVED***;
        function isAncestorNamespaceDeclared(node) {
            let ancestor = node.parent;
            while (ancestor) {
                if (ancestor.type === utils_1.AST_NODE_TYPES.TSModuleDeclaration &&
                    ancestor.declare) {
                    return true;
                ***REMOVED***
                ancestor = ancestor.parent;
            ***REMOVED***
            return false;
        ***REMOVED***
    ***REMOVED***,
***REMOVED***);
