// @typescript-eslint/eslint-plugin/dist/rules/member-delimiter-style.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; ***REMOVED*** ***REMOVED***;
    ***REMOVED***
    Object.defineProperty(o, k2, desc);
***REMOVED***) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
***REMOVED***));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v ***REMOVED***);
***REMOVED***) : function(o, v) {
    o["default"] = v;
***REMOVED***);
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {***REMOVED***;
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const utils_1 = require("@typescript-eslint/utils");
const util = __importStar(require("../util"));
const definition = {
    type: 'object',
    properties: {
        multiline: {
            type: 'object',
            properties: {
                delimiter: { enum: ['none', 'semi', 'comma'] ***REMOVED***,
                requireLast: { type: 'boolean' ***REMOVED***,
            ***REMOVED***,
            additionalProperties: false,
        ***REMOVED***,
        singleline: {
            type: 'object',
            properties: {
                // note can't have "none" for single line delimiter as it's invalid syntax
                delimiter: { enum: ['semi', 'comma'] ***REMOVED***,
                requireLast: { type: 'boolean' ***REMOVED***,
            ***REMOVED***,
            additionalProperties: false,
        ***REMOVED***,
    ***REMOVED***,
    additionalProperties: false,
***REMOVED***;
const isLastTokenEndOfLine = (token, line) => {
    const positionInLine = token.loc.start.column;
    return positionInLine === line.length - 1;
***REMOVED***;
const isCommentsEndOfLine = (token, comments, line) => {
    if (!comments) {
        return false;
    ***REMOVED***
    if (comments.loc.end.line > token.loc.end.line) {
        return true;
    ***REMOVED***
    const positionInLine = comments.loc.end.column;
    return positionInLine === line.length;
***REMOVED***;
const makeFixFunction = ({ optsNone, optsSemi, lastToken, commentsAfterLastToken, missingDelimiter, lastTokenLine, isSingleLine, ***REMOVED***) => {
    // if removing is the action but last token is not the end of the line
    if (optsNone &&
        !isLastTokenEndOfLine(lastToken, lastTokenLine) &&
        !isCommentsEndOfLine(lastToken, commentsAfterLastToken, lastTokenLine) &&
        !isSingleLine) {
        return null;
    ***REMOVED***
    return (fixer) => {
        if (optsNone) {
            // remove the unneeded token
            return fixer.remove(lastToken);
        ***REMOVED***
        const token = optsSemi ? ';' : ',';
        if (missingDelimiter) {
            // add the missing delimiter
            return fixer.insertTextAfter(lastToken, token);
        ***REMOVED***
        // correct the current delimiter
        return fixer.replaceText(lastToken, token);
    ***REMOVED***;
***REMOVED***;
exports.default = util.createRule({
    name: 'member-delimiter-style',
    meta: {
        type: 'layout',
        docs: {
            description: 'Require a specific member delimiter style for interfaces and type literals',
            recommended: false,
        ***REMOVED***,
        fixable: 'whitespace',
        messages: {
            unexpectedComma: 'Unexpected separator (,).',
            unexpectedSemi: 'Unexpected separator (;).',
            expectedComma: 'Expected a comma.',
            expectedSemi: 'Expected a semicolon.',
        ***REMOVED***,
        schema: [
            {
                type: 'object',
                properties: Object.assign({***REMOVED***, definition.properties, {
                    overrides: {
                        type: 'object',
                        properties: {
                            interface: definition,
                            typeLiteral: definition,
                        ***REMOVED***,
                        additionalProperties: false,
                    ***REMOVED***,
                    multilineDetection: {
                        enum: ['brackets', 'last-member'],
                    ***REMOVED***,
                ***REMOVED***),
                additionalProperties: false,
            ***REMOVED***,
        ],
    ***REMOVED***,
    defaultOptions: [
        {
            multiline: {
                delimiter: 'semi',
                requireLast: true,
            ***REMOVED***,
            singleline: {
                delimiter: 'semi',
                requireLast: false,
            ***REMOVED***,
            multilineDetection: 'brackets',
        ***REMOVED***,
    ],
    create(context, [options]) {
        var _a;
        const sourceCode = context.getSourceCode();
        // use the base options as the defaults for the cases
        const baseOptions = options;
        const overrides = (_a = baseOptions.overrides) !== null && _a !== void 0 ? _a : {***REMOVED***;
        const interfaceOptions = util.deepMerge(baseOptions, overrides.interface);
        const typeLiteralOptions = util.deepMerge(baseOptions, overrides.typeLiteral);
        /**
         * Check the last token in the given member.
         * @param member the member to be evaluated.
         * @param opts the options to be validated.
         * @param isLast a flag indicating `member` is the last in the interface or type literal.
         */
        function checkLastToken(member, opts, isLast) {
            /**
             * Resolves the boolean value for the given setting enum value
             * @param type the option name
             */
            function getOption(type) {
                if (isLast && !opts.requireLast) {
                    // only turn the option on if its expecting no delimiter for the last member
                    return type === 'none';
                ***REMOVED***
                return opts.delimiter === type;
            ***REMOVED***
            let messageId = null;
            let missingDelimiter = false;
            const lastToken = sourceCode.getLastToken(member, {
                includeComments: false,
            ***REMOVED***);
            if (!lastToken) {
                return;
            ***REMOVED***
            const commentsAfterLastToken = sourceCode
                .getCommentsAfter(lastToken)
                .pop();
            const sourceCodeLines = sourceCode.getLines();
            const lastTokenLine = sourceCodeLines[(lastToken === null || lastToken === void 0 ? void 0 : lastToken.loc.start.line) - 1];
            const optsSemi = getOption('semi');
            const optsComma = getOption('comma');
            const optsNone = getOption('none');
            if (lastToken.value === ';') {
                if (optsComma) {
                    messageId = 'expectedComma';
                ***REMOVED***
                else if (optsNone) {
                    missingDelimiter = true;
                    messageId = 'unexpectedSemi';
                ***REMOVED***
            ***REMOVED***
            else if (lastToken.value === ',') {
                if (optsSemi) {
                    messageId = 'expectedSemi';
                ***REMOVED***
                else if (optsNone) {
                    missingDelimiter = true;
                    messageId = 'unexpectedComma';
                ***REMOVED***
            ***REMOVED***
            else {
                if (optsSemi) {
                    missingDelimiter = true;
                    messageId = 'expectedSemi';
                ***REMOVED***
                else if (optsComma) {
                    missingDelimiter = true;
                    messageId = 'expectedComma';
                ***REMOVED***
            ***REMOVED***
            if (messageId) {
                context.report({
                    node: lastToken,
                    loc: {
                        start: {
                            line: lastToken.loc.end.line,
                            column: lastToken.loc.end.column,
                        ***REMOVED***,
                        end: {
                            line: lastToken.loc.end.line,
                            column: lastToken.loc.end.column,
                        ***REMOVED***,
                    ***REMOVED***,
                    messageId,
                    fix: makeFixFunction({
                        optsNone,
                        optsSemi,
                        lastToken,
                        commentsAfterLastToken,
                        missingDelimiter,
                        lastTokenLine,
                        isSingleLine: opts.type === 'single-line',
                    ***REMOVED***),
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***
        /**
         * Check the member separator being used matches the delimiter.
         * @param {ASTNode***REMOVED*** node the node to be evaluated.
         */
        function checkMemberSeparatorStyle(node) {
            const members = node.type === utils_1.AST_NODE_TYPES.TSInterfaceBody ? node.body : node.members;
            let isSingleLine = node.loc.start.line === node.loc.end.line;
            if (options.multilineDetection === 'last-member' &&
                !isSingleLine &&
                members.length > 0) {
                const lastMember = members[members.length - 1];
                if (lastMember.loc.end.line === node.loc.end.line) {
                    isSingleLine = true;
                ***REMOVED***
            ***REMOVED***
            const typeOpts = node.type === utils_1.AST_NODE_TYPES.TSInterfaceBody
                ? interfaceOptions
                : typeLiteralOptions;
            const opts = isSingleLine
                ? Object.assign(Object.assign({***REMOVED***, typeOpts.singleline), { type: 'single-line' ***REMOVED***) : Object.assign(Object.assign({***REMOVED***, typeOpts.multiline), { type: 'multi-line' ***REMOVED***);
            members.forEach((member, index) => {
                checkLastToken(member, opts !== null && opts !== void 0 ? opts : {***REMOVED***, index === members.length - 1);
            ***REMOVED***);
        ***REMOVED***
        return {
            TSInterfaceBody: checkMemberSeparatorStyle,
            TSTypeLiteral: checkMemberSeparatorStyle,
        ***REMOVED***;
    ***REMOVED***,
***REMOVED***);
