// @typescript-eslint/eslint-plugin/dist/rules/naming-convention.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; ***REMOVED*** ***REMOVED***;
    ***REMOVED***
    Object.defineProperty(o, k2, desc);
***REMOVED***) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
***REMOVED***));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v ***REMOVED***);
***REMOVED***) : function(o, v) {
    o["default"] = v;
***REMOVED***);
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {***REMOVED***;
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const scope_manager_1 = require("@typescript-eslint/scope-manager");
const utils_1 = require("@typescript-eslint/utils");
const util = __importStar(require("../util"));
const naming_convention_utils_1 = require("./naming-convention-utils");
// This essentially mirrors ESLint's `camelcase` rule
// note that that rule ignores leading and trailing underscores and only checks those in the middle of a variable name
const defaultCamelCaseAllTheThingsConfig = [
    {
        selector: 'default',
        format: ['camelCase'],
        leadingUnderscore: 'allow',
        trailingUnderscore: 'allow',
    ***REMOVED***,
    {
        selector: 'variable',
        format: ['camelCase', 'UPPER_CASE'],
        leadingUnderscore: 'allow',
        trailingUnderscore: 'allow',
    ***REMOVED***,
    {
        selector: 'typeLike',
        format: ['PascalCase'],
    ***REMOVED***,
];
exports.default = util.createRule({
    name: 'naming-convention',
    meta: {
        docs: {
            description: 'Enforce naming conventions for everything across a codebase',
            recommended: false,
            // technically only requires type checking if the user uses "type" modifiers
            requiresTypeChecking: true,
        ***REMOVED***,
        type: 'suggestion',
        messages: {
            unexpectedUnderscore: '{{type***REMOVED******REMOVED*** name `{{name***REMOVED******REMOVED***` must not have a {{position***REMOVED******REMOVED*** underscore.',
            missingUnderscore: '{{type***REMOVED******REMOVED*** name `{{name***REMOVED******REMOVED***` must have {{count***REMOVED******REMOVED*** {{position***REMOVED******REMOVED*** underscore(s).',
            missingAffix: '{{type***REMOVED******REMOVED*** name `{{name***REMOVED******REMOVED***` must have one of the following {{position***REMOVED******REMOVED***es: {{affixes***REMOVED******REMOVED***',
            satisfyCustom: '{{type***REMOVED******REMOVED*** name `{{name***REMOVED******REMOVED***` must {{regexMatch***REMOVED******REMOVED*** the RegExp: {{regex***REMOVED******REMOVED***',
            doesNotMatchFormat: '{{type***REMOVED******REMOVED*** name `{{name***REMOVED******REMOVED***` must match one of the following formats: {{formats***REMOVED******REMOVED***',
            doesNotMatchFormatTrimmed: '{{type***REMOVED******REMOVED*** name `{{name***REMOVED******REMOVED***` trimmed as `{{processedName***REMOVED******REMOVED***` must match one of the following formats: {{formats***REMOVED******REMOVED***',
        ***REMOVED***,
        schema: naming_convention_utils_1.SCHEMA,
    ***REMOVED***,
    defaultOptions: defaultCamelCaseAllTheThingsConfig,
    create(contextWithoutDefaults) {
        const context = contextWithoutDefaults.options &&
            contextWithoutDefaults.options.length > 0
            ? contextWithoutDefaults
            : // only apply the defaults when the user provides no config
                Object.setPrototypeOf({
                    options: defaultCamelCaseAllTheThingsConfig,
                ***REMOVED***, contextWithoutDefaults);
        const validators = (0, naming_convention_utils_1.parseOptions)(context);
        // getParserServices(context, false) -- dirty hack to work around the docs checker test...
        const compilerOptions = util
            .getParserServices(context, true)
            .program.getCompilerOptions();
        function handleMember(validator, node, modifiers) {
            if (!validator) {
                return;
            ***REMOVED***
            const key = node.key;
            if (requiresQuoting(key, compilerOptions.target)) {
                modifiers.add(naming_convention_utils_1.Modifiers.requiresQuotes);
            ***REMOVED***
            validator(key, modifiers);
        ***REMOVED***
        function getMemberModifiers(node) {
            const modifiers = new Set();
            if ('key' in node && node.key.type === utils_1.AST_NODE_TYPES.PrivateIdentifier) {
                modifiers.add(naming_convention_utils_1.Modifiers['#private']);
            ***REMOVED***
            else if (node.accessibility) {
                modifiers.add(naming_convention_utils_1.Modifiers[node.accessibility]);
            ***REMOVED***
            else {
                modifiers.add(naming_convention_utils_1.Modifiers.public);
            ***REMOVED***
            if (node.static) {
                modifiers.add(naming_convention_utils_1.Modifiers.static);
            ***REMOVED***
            if ('readonly' in node && node.readonly) {
                modifiers.add(naming_convention_utils_1.Modifiers.readonly);
            ***REMOVED***
            if ('override' in node && node.override) {
                modifiers.add(naming_convention_utils_1.Modifiers.override);
            ***REMOVED***
            if (node.type === utils_1.AST_NODE_TYPES.TSAbstractPropertyDefinition ||
                node.type === utils_1.AST_NODE_TYPES.TSAbstractMethodDefinition) {
                modifiers.add(naming_convention_utils_1.Modifiers.abstract);
            ***REMOVED***
            return modifiers;
        ***REMOVED***
        const unusedVariables = util.collectUnusedVariables(context);
        function isUnused(name, initialScope = context.getScope()) {
            var _a;
            let variable = null;
            let scope = initialScope;
            while (scope) {
                variable = (_a = scope.set.get(name)) !== null && _a !== void 0 ? _a : null;
                if (variable) {
                    break;
                ***REMOVED***
                scope = scope.upper;
            ***REMOVED***
            if (!variable) {
                return false;
            ***REMOVED***
            return unusedVariables.has(variable);
        ***REMOVED***
        function isDestructured(id) {
            var _a, _b, _c;
            return (
            // `const { x ***REMOVED***`
            // does not match `const { x: y ***REMOVED***`
            (((_a = id.parent) === null || _a === void 0 ? void 0 : _a.type) === utils_1.AST_NODE_TYPES.Property && id.parent.shorthand) ||
                // `const { x = 2 ***REMOVED***`
                // does not match const `{ x: y = 2 ***REMOVED***`
                (((_b = id.parent) === null || _b === void 0 ? void 0 : _b.type) === utils_1.AST_NODE_TYPES.AssignmentPattern &&
                    ((_c = id.parent.parent) === null || _c === void 0 ? void 0 : _c.type) === utils_1.AST_NODE_TYPES.Property &&
                    id.parent.parent.shorthand));
        ***REMOVED***
        function isAsyncMemberOrProperty(propertyOrMemberNode) {
            return Boolean('value' in propertyOrMemberNode &&
                propertyOrMemberNode.value &&
                'async' in propertyOrMemberNode.value &&
                propertyOrMemberNode.value.async);
        ***REMOVED***
        function isAsyncVariableIdentifier(id) {
            return Boolean(id.parent &&
                (('async' in id.parent && id.parent.async) ||
                    ('init' in id.parent &&
                        id.parent.init &&
                        'async' in id.parent.init &&
                        id.parent.init.async)));
        ***REMOVED***
        const selectors = {
            // #region variable
            VariableDeclarator: {
                validator: validators.variable,
                handler: (node, validator) => {
                    const identifiers = getIdentifiersFromPattern(node.id);
                    const baseModifiers = new Set();
                    const parent = node.parent;
                    if ((parent === null || parent === void 0 ? void 0 : parent.type) === utils_1.AST_NODE_TYPES.VariableDeclaration) {
                        if (parent.kind === 'const') {
                            baseModifiers.add(naming_convention_utils_1.Modifiers.const);
                        ***REMOVED***
                        if (isGlobal(context.getScope())) {
                            baseModifiers.add(naming_convention_utils_1.Modifiers.global);
                        ***REMOVED***
                    ***REMOVED***
                    identifiers.forEach(id => {
                        const modifiers = new Set(baseModifiers);
                        if (isDestructured(id)) {
                            modifiers.add(naming_convention_utils_1.Modifiers.destructured);
                        ***REMOVED***
                        if (isExported(parent, id.name, context.getScope())) {
                            modifiers.add(naming_convention_utils_1.Modifiers.exported);
                        ***REMOVED***
                        if (isUnused(id.name)) {
                            modifiers.add(naming_convention_utils_1.Modifiers.unused);
                        ***REMOVED***
                        if (isAsyncVariableIdentifier(id)) {
                            modifiers.add(naming_convention_utils_1.Modifiers.async);
                        ***REMOVED***
                        validator(id, modifiers);
                    ***REMOVED***);
                ***REMOVED***,
            ***REMOVED***,
            // #endregion
            // #region function
            'FunctionDeclaration, TSDeclareFunction, FunctionExpression': {
                validator: validators.function,
                handler: (node, validator) => {
                    if (node.id == null) {
                        return;
                    ***REMOVED***
                    const modifiers = new Set();
                    // functions create their own nested scope
                    const scope = context.getScope().upper;
                    if (isGlobal(scope)) {
                        modifiers.add(naming_convention_utils_1.Modifiers.global);
                    ***REMOVED***
                    if (isExported(node, node.id.name, scope)) {
                        modifiers.add(naming_convention_utils_1.Modifiers.exported);
                    ***REMOVED***
                    if (isUnused(node.id.name, scope)) {
                        modifiers.add(naming_convention_utils_1.Modifiers.unused);
                    ***REMOVED***
                    if (node.async) {
                        modifiers.add(naming_convention_utils_1.Modifiers.async);
                    ***REMOVED***
                    validator(node.id, modifiers);
                ***REMOVED***,
            ***REMOVED***,
            // #endregion function
            // #region parameter
            'FunctionDeclaration, TSDeclareFunction, TSEmptyBodyFunctionExpression, FunctionExpression, ArrowFunctionExpression': {
                validator: validators.parameter,
                handler: (node, validator) => {
                    node.params.forEach(param => {
                        if (param.type === utils_1.AST_NODE_TYPES.TSParameterProperty) {
                            return;
                        ***REMOVED***
                        const identifiers = getIdentifiersFromPattern(param);
                        identifiers.forEach(i => {
                            const modifiers = new Set();
                            if (isDestructured(i)) {
                                modifiers.add(naming_convention_utils_1.Modifiers.destructured);
                            ***REMOVED***
                            if (isUnused(i.name)) {
                                modifiers.add(naming_convention_utils_1.Modifiers.unused);
                            ***REMOVED***
                            validator(i, modifiers);
                        ***REMOVED***);
                    ***REMOVED***);
                ***REMOVED***,
            ***REMOVED***,
            // #endregion parameter
            // #region parameterProperty
            TSParameterProperty: {
                validator: validators.parameterProperty,
                handler: (node, validator) => {
                    const modifiers = getMemberModifiers(node);
                    const identifiers = getIdentifiersFromPattern(node.parameter);
                    identifiers.forEach(i => {
                        validator(i, modifiers);
                    ***REMOVED***);
                ***REMOVED***,
            ***REMOVED***,
            // #endregion parameterProperty
            // #region property
            ':not(ObjectPattern) > Property[computed = false][kind = "init"][value.type != "ArrowFunctionExpression"][value.type != "FunctionExpression"][value.type != "TSEmptyBodyFunctionExpression"]': {
                validator: validators.objectLiteralProperty,
                handler: (node, validator) => {
                    const modifiers = new Set([naming_convention_utils_1.Modifiers.public]);
                    handleMember(validator, node, modifiers);
                ***REMOVED***,
            ***REMOVED***,
            ':matches(PropertyDefinition, TSAbstractPropertyDefinition)[computed = false][value.type != "ArrowFunctionExpression"][value.type != "FunctionExpression"][value.type != "TSEmptyBodyFunctionExpression"]': {
                validator: validators.classProperty,
                handler: (node, validator) => {
                    const modifiers = getMemberModifiers(node);
                    handleMember(validator, node, modifiers);
                ***REMOVED***,
            ***REMOVED***,
            'TSPropertySignature[computed = false][typeAnnotation.typeAnnotation.type != "TSFunctionType"]': {
                validator: validators.typeProperty,
                handler: (node, validator) => {
                    const modifiers = new Set([naming_convention_utils_1.Modifiers.public]);
                    if (node.readonly) {
                        modifiers.add(naming_convention_utils_1.Modifiers.readonly);
                    ***REMOVED***
                    handleMember(validator, node, modifiers);
                ***REMOVED***,
            ***REMOVED***,
            // #endregion property
            // #region method
            [[
                'Property[computed = false][kind = "init"][value.type = "ArrowFunctionExpression"]',
                'Property[computed = false][kind = "init"][value.type = "FunctionExpression"]',
                'Property[computed = false][kind = "init"][value.type = "TSEmptyBodyFunctionExpression"]',
            ].join(', ')]: {
                validator: validators.objectLiteralMethod,
                handler: (node, validator) => {
                    const modifiers = new Set([naming_convention_utils_1.Modifiers.public]);
                    if (isAsyncMemberOrProperty(node)) {
                        modifiers.add(naming_convention_utils_1.Modifiers.async);
                    ***REMOVED***
                    handleMember(validator, node, modifiers);
                ***REMOVED***,
            ***REMOVED***,
            [[
                ':matches(PropertyDefinition, TSAbstractPropertyDefinition)[computed = false][value.type = "ArrowFunctionExpression"]',
                ':matches(PropertyDefinition, TSAbstractPropertyDefinition)[computed = false][value.type = "FunctionExpression"]',
                ':matches(PropertyDefinition, TSAbstractPropertyDefinition)[computed = false][value.type = "TSEmptyBodyFunctionExpression"]',
                ':matches(MethodDefinition, TSAbstractMethodDefinition)[computed = false][kind = "method"]',
            ].join(', ')]: {
                validator: validators.classMethod,
                handler: (node, validator) => {
                    const modifiers = getMemberModifiers(node);
                    if (isAsyncMemberOrProperty(node)) {
                        modifiers.add(naming_convention_utils_1.Modifiers.async);
                    ***REMOVED***
                    handleMember(validator, node, modifiers);
                ***REMOVED***,
            ***REMOVED***,
            [[
                'TSMethodSignature[computed = false]',
                'TSPropertySignature[computed = false][typeAnnotation.typeAnnotation.type = "TSFunctionType"]',
            ].join(', ')]: {
                validator: validators.typeMethod,
                handler: (node, validator) => {
                    const modifiers = new Set([naming_convention_utils_1.Modifiers.public]);
                    handleMember(validator, node, modifiers);
                ***REMOVED***,
            ***REMOVED***,
            // #endregion method
            // #region accessor
            'Property[computed = false]:matches([kind = "get"], [kind = "set"])': {
                validator: validators.accessor,
                handler: (node, validator) => {
                    const modifiers = new Set([naming_convention_utils_1.Modifiers.public]);
                    handleMember(validator, node, modifiers);
                ***REMOVED***,
            ***REMOVED***,
            'MethodDefinition[computed = false]:matches([kind = "get"], [kind = "set"])': {
                validator: validators.accessor,
                handler: (node, validator) => {
                    const modifiers = getMemberModifiers(node);
                    handleMember(validator, node, modifiers);
                ***REMOVED***,
            ***REMOVED***,
            // #endregion accessor
            // #region enumMember
            // computed is optional, so can't do [computed = false]
            'TSEnumMember[computed != true]': {
                validator: validators.enumMember,
                handler: (node, validator) => {
                    const id = node.id;
                    const modifiers = new Set();
                    if (requiresQuoting(id, compilerOptions.target)) {
                        modifiers.add(naming_convention_utils_1.Modifiers.requiresQuotes);
                    ***REMOVED***
                    validator(id, modifiers);
                ***REMOVED***,
            ***REMOVED***,
            // #endregion enumMember
            // #region class
            'ClassDeclaration, ClassExpression': {
                validator: validators.class,
                handler: (node, validator) => {
                    const id = node.id;
                    if (id == null) {
                        return;
                    ***REMOVED***
                    const modifiers = new Set();
                    // classes create their own nested scope
                    const scope = context.getScope().upper;
                    if (node.abstract) {
                        modifiers.add(naming_convention_utils_1.Modifiers.abstract);
                    ***REMOVED***
                    if (isExported(node, id.name, scope)) {
                        modifiers.add(naming_convention_utils_1.Modifiers.exported);
                    ***REMOVED***
                    if (isUnused(id.name, scope)) {
                        modifiers.add(naming_convention_utils_1.Modifiers.unused);
                    ***REMOVED***
                    validator(id, modifiers);
                ***REMOVED***,
            ***REMOVED***,
            // #endregion class
            // #region interface
            TSInterfaceDeclaration: {
                validator: validators.interface,
                handler: (node, validator) => {
                    const modifiers = new Set();
                    const scope = context.getScope();
                    if (isExported(node, node.id.name, scope)) {
                        modifiers.add(naming_convention_utils_1.Modifiers.exported);
                    ***REMOVED***
                    if (isUnused(node.id.name, scope)) {
                        modifiers.add(naming_convention_utils_1.Modifiers.unused);
                    ***REMOVED***
                    validator(node.id, modifiers);
                ***REMOVED***,
            ***REMOVED***,
            // #endregion interface
            // #region typeAlias
            TSTypeAliasDeclaration: {
                validator: validators.typeAlias,
                handler: (node, validator) => {
                    const modifiers = new Set();
                    const scope = context.getScope();
                    if (isExported(node, node.id.name, scope)) {
                        modifiers.add(naming_convention_utils_1.Modifiers.exported);
                    ***REMOVED***
                    if (isUnused(node.id.name, scope)) {
                        modifiers.add(naming_convention_utils_1.Modifiers.unused);
                    ***REMOVED***
                    validator(node.id, modifiers);
                ***REMOVED***,
            ***REMOVED***,
            // #endregion typeAlias
            // #region enum
            TSEnumDeclaration: {
                validator: validators.enum,
                handler: (node, validator) => {
                    const modifiers = new Set();
                    // enums create their own nested scope
                    const scope = context.getScope().upper;
                    if (isExported(node, node.id.name, scope)) {
                        modifiers.add(naming_convention_utils_1.Modifiers.exported);
                    ***REMOVED***
                    if (isUnused(node.id.name, scope)) {
                        modifiers.add(naming_convention_utils_1.Modifiers.unused);
                    ***REMOVED***
                    validator(node.id, modifiers);
                ***REMOVED***,
            ***REMOVED***,
            // #endregion enum
            // #region typeParameter
            'TSTypeParameterDeclaration > TSTypeParameter': {
                validator: validators.typeParameter,
                handler: (node, validator) => {
                    const modifiers = new Set();
                    const scope = context.getScope();
                    if (isUnused(node.name.name, scope)) {
                        modifiers.add(naming_convention_utils_1.Modifiers.unused);
                    ***REMOVED***
                    validator(node.name, modifiers);
                ***REMOVED***,
            ***REMOVED***,
            // #endregion typeParameter
        ***REMOVED***;
        return Object.fromEntries(Object.entries(selectors)
            .map(([selector, { validator, handler ***REMOVED***]) => {
            return [
                selector,
                (node) => {
                    handler(node, validator);
                ***REMOVED***,
            ];
        ***REMOVED***)
            .filter((s) => s != null));
    ***REMOVED***,
***REMOVED***);
function getIdentifiersFromPattern(pattern) {
    const identifiers = [];
    const visitor = new scope_manager_1.PatternVisitor({***REMOVED***, pattern, id => identifiers.push(id));
    visitor.visit(pattern);
    return identifiers;
***REMOVED***
function isExported(node, name, scope) {
    var _a, _b;
    if (((_a = node === null || node === void 0 ? void 0 : node.parent) === null || _a === void 0 ? void 0 : _a.type) === utils_1.AST_NODE_TYPES.ExportDefaultDeclaration ||
        ((_b = node === null || node === void 0 ? void 0 : node.parent) === null || _b === void 0 ? void 0 : _b.type) === utils_1.AST_NODE_TYPES.ExportNamedDeclaration) {
        return true;
    ***REMOVED***
    if (scope == null) {
        return false;
    ***REMOVED***
    const variable = scope.set.get(name);
    if (variable) {
        for (const ref of variable.references) {
            const refParent = ref.identifier.parent;
            if ((refParent === null || refParent === void 0 ? void 0 : refParent.type) === utils_1.AST_NODE_TYPES.ExportDefaultDeclaration ||
                (refParent === null || refParent === void 0 ? void 0 : refParent.type) === utils_1.AST_NODE_TYPES.ExportSpecifier) {
                return true;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    return false;
***REMOVED***
function isGlobal(scope) {
    if (scope == null) {
        return false;
    ***REMOVED***
    return (scope.type === utils_1.TSESLint.Scope.ScopeType.global ||
        scope.type === utils_1.TSESLint.Scope.ScopeType.module);
***REMOVED***
function requiresQuoting(node, target) {
    const name = node.type === utils_1.AST_NODE_TYPES.Identifier ||
        node.type === utils_1.AST_NODE_TYPES.PrivateIdentifier
        ? node.name
        : `${node.value***REMOVED***`;
    return util.requiresQuoting(name, target);
***REMOVED***
