// @typescript-eslint/eslint-plugin/dist/rules/no-base-to-string.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; ***REMOVED*** ***REMOVED***;
    ***REMOVED***
    Object.defineProperty(o, k2, desc);
***REMOVED***) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
***REMOVED***));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v ***REMOVED***);
***REMOVED***) : function(o, v) {
    o["default"] = v;
***REMOVED***);
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {***REMOVED***;
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const utils_1 = require("@typescript-eslint/utils");
const ts = __importStar(require("typescript"));
const util = __importStar(require("../util"));
var Usefulness;
(function (Usefulness) {
    Usefulness[Usefulness["Always"] = 0] = "Always";
    Usefulness["Never"] = "will";
    Usefulness["Sometimes"] = "may";
***REMOVED***)(Usefulness || (Usefulness = {***REMOVED***));
exports.default = util.createRule({
    name: 'no-base-to-string',
    meta: {
        docs: {
            description: 'Require `.toString()` to only be called on objects which provide useful information when stringified',
            recommended: 'strict',
            requiresTypeChecking: true,
        ***REMOVED***,
        messages: {
            baseToString: "'{{name***REMOVED******REMOVED***' {{certainty***REMOVED******REMOVED*** evaluate to '[object Object]' when stringified.",
        ***REMOVED***,
        schema: [
            {
                type: 'object',
                properties: {
                    ignoredTypeNames: {
                        type: 'array',
                        items: {
                            type: 'string',
                        ***REMOVED***,
                    ***REMOVED***,
                ***REMOVED***,
                additionalProperties: false,
            ***REMOVED***,
        ],
        type: 'suggestion',
    ***REMOVED***,
    defaultOptions: [
        {
            ignoredTypeNames: ['Error', 'RegExp', 'URL', 'URLSearchParams'],
        ***REMOVED***,
    ],
    create(context, [option]) {
        var _a;
        const parserServices = util.getParserServices(context);
        const typeChecker = parserServices.program.getTypeChecker();
        const ignoredTypeNames = (_a = option.ignoredTypeNames) !== null && _a !== void 0 ? _a : [];
        function checkExpression(node, type) {
            if (node.type === utils_1.AST_NODE_TYPES.Literal) {
                return;
            ***REMOVED***
            const certainty = collectToStringCertainty(type !== null && type !== void 0 ? type : typeChecker.getTypeAtLocation(parserServices.esTreeNodeToTSNodeMap.get(node)));
            if (certainty === Usefulness.Always) {
                return;
            ***REMOVED***
            context.report({
                data: {
                    certainty,
                    name: context.getSourceCode().getText(node),
                ***REMOVED***,
                messageId: 'baseToString',
                node,
            ***REMOVED***);
        ***REMOVED***
        function collectToStringCertainty(type) {
            const toString = typeChecker.getPropertyOfType(type, 'toString');
            const declarations = toString === null || toString === void 0 ? void 0 : toString.getDeclarations();
            if (!toString || !declarations || declarations.length === 0) {
                return Usefulness.Always;
            ***REMOVED***
            // Patch for old version TypeScript, the Boolean type definition missing toString()
            if (type.flags & ts.TypeFlags.Boolean ||
                type.flags & ts.TypeFlags.BooleanLiteral) {
                return Usefulness.Always;
            ***REMOVED***
            if (ignoredTypeNames.includes(util.getTypeName(typeChecker, type))) {
                return Usefulness.Always;
            ***REMOVED***
            if (declarations.every(({ parent ***REMOVED***) => !ts.isInterfaceDeclaration(parent) || parent.name.text !== 'Object')) {
                return Usefulness.Always;
            ***REMOVED***
            if (type.isIntersection()) {
                for (const subType of type.types) {
                    const subtypeUsefulness = collectToStringCertainty(subType);
                    if (subtypeUsefulness === Usefulness.Always) {
                        return Usefulness.Always;
                    ***REMOVED***
                ***REMOVED***
                return Usefulness.Never;
            ***REMOVED***
            if (!type.isUnion()) {
                return Usefulness.Never;
            ***REMOVED***
            let allSubtypesUseful = true;
            let someSubtypeUseful = false;
            for (const subType of type.types) {
                const subtypeUsefulness = collectToStringCertainty(subType);
                if (subtypeUsefulness !== Usefulness.Always && allSubtypesUseful) {
                    allSubtypesUseful = false;
                ***REMOVED***
                if (subtypeUsefulness !== Usefulness.Never && !someSubtypeUseful) {
                    someSubtypeUseful = true;
                ***REMOVED***
            ***REMOVED***
            if (allSubtypesUseful && someSubtypeUseful) {
                return Usefulness.Always;
            ***REMOVED***
            if (someSubtypeUseful) {
                return Usefulness.Sometimes;
            ***REMOVED***
            return Usefulness.Never;
        ***REMOVED***
        return {
            'AssignmentExpression[operator = "+="], BinaryExpression[operator = "+"]'(node) {
                const leftType = typeChecker.getTypeAtLocation(parserServices.esTreeNodeToTSNodeMap.get(node.left));
                const rightType = typeChecker.getTypeAtLocation(parserServices.esTreeNodeToTSNodeMap.get(node.right));
                if (util.getTypeName(typeChecker, leftType) === 'string') {
                    checkExpression(node.right, rightType);
                ***REMOVED***
                else if (util.getTypeName(typeChecker, rightType) === 'string' &&
                    node.left.type !== utils_1.AST_NODE_TYPES.PrivateIdentifier) {
                    checkExpression(node.left, leftType);
                ***REMOVED***
            ***REMOVED***,
            'CallExpression > MemberExpression.callee > Identifier[name = "toString"].property'(node) {
                const memberExpr = node.parent;
                checkExpression(memberExpr.object);
            ***REMOVED***,
            TemplateLiteral(node) {
                if (node.parent &&
                    node.parent.type === utils_1.AST_NODE_TYPES.TaggedTemplateExpression) {
                    return;
                ***REMOVED***
                for (const expression of node.expressions) {
                    checkExpression(expression);
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***;
    ***REMOVED***,
***REMOVED***);
