// @typescript-eslint/eslint-plugin/dist/rules/no-duplicate-type-constituents.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; ***REMOVED*** ***REMOVED***;
    ***REMOVED***
    Object.defineProperty(o, k2, desc);
***REMOVED***) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
***REMOVED***));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v ***REMOVED***);
***REMOVED***) : function(o, v) {
    o["default"] = v;
***REMOVED***);
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {***REMOVED***;
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const utils_1 = require("@typescript-eslint/utils");
const util = __importStar(require("../util"));
const astIgnoreKeys = new Set(['range', 'loc', 'parent']);
const isSameAstNode = (actualNode, expectedNode) => {
    if (actualNode === expectedNode) {
        return true;
    ***REMOVED***
    if (actualNode &&
        expectedNode &&
        typeof actualNode === 'object' &&
        typeof expectedNode === 'object') {
        if (Array.isArray(actualNode) && Array.isArray(expectedNode)) {
            if (actualNode.length !== expectedNode.length) {
                return false;
            ***REMOVED***
            return !actualNode.some((nodeEle, index) => !isSameAstNode(nodeEle, expectedNode[index]));
        ***REMOVED***
        const actualNodeKeys = Object.keys(actualNode).filter(key => !astIgnoreKeys.has(key));
        const expectedNodeKeys = Object.keys(expectedNode).filter(key => !astIgnoreKeys.has(key));
        if (actualNodeKeys.length !== expectedNodeKeys.length) {
            return false;
        ***REMOVED***
        if (actualNodeKeys.some(actualNodeKey => !Object.prototype.hasOwnProperty.call(expectedNode, actualNodeKey))) {
            return false;
        ***REMOVED***
        if (actualNodeKeys.some(actualNodeKey => !isSameAstNode(actualNode[actualNodeKey], expectedNode[actualNodeKey]))) {
            return false;
        ***REMOVED***
        return true;
    ***REMOVED***
    return false;
***REMOVED***;
exports.default = util.createRule({
    name: 'no-duplicate-type-constituents',
    meta: {
        type: 'suggestion',
        docs: {
            description: 'Disallow duplicate constituents of union or intersection types',
            recommended: false,
            requiresTypeChecking: true,
        ***REMOVED***,
        fixable: 'code',
        messages: {
            duplicate: '{{type***REMOVED******REMOVED*** type constituent is duplicated with {{previous***REMOVED******REMOVED***.',
        ***REMOVED***,
        schema: [
            {
                additionalProperties: false,
                type: 'object',
                properties: {
                    ignoreIntersections: {
                        type: 'boolean',
                    ***REMOVED***,
                    ignoreUnions: {
                        type: 'boolean',
                    ***REMOVED***,
                ***REMOVED***,
            ***REMOVED***,
        ],
    ***REMOVED***,
    defaultOptions: [
        {
            ignoreIntersections: false,
            ignoreUnions: false,
        ***REMOVED***,
    ],
    create(context, [{ ignoreIntersections, ignoreUnions ***REMOVED***]) {
        const parserServices = util.getParserServices(context);
        const checker = parserServices.program.getTypeChecker();
        function checkDuplicate(node) {
            const cachedTypeMap = new Map();
            node.types.reduce((uniqueConstituents, constituentNode) => {
                const duplicatedPreviousConstituentInAst = uniqueConstituents.find(ele => isSameAstNode(ele, constituentNode));
                if (duplicatedPreviousConstituentInAst) {
                    reportDuplicate({
                        duplicated: constituentNode,
                        duplicatePrevious: duplicatedPreviousConstituentInAst,
                    ***REMOVED***, node);
                    return uniqueConstituents;
                ***REMOVED***
                const constituentNodeType = checker.getTypeAtLocation(parserServices.esTreeNodeToTSNodeMap.get(constituentNode));
                const duplicatedPreviousConstituentInType = cachedTypeMap.get(constituentNodeType);
                if (duplicatedPreviousConstituentInType) {
                    reportDuplicate({
                        duplicated: constituentNode,
                        duplicatePrevious: duplicatedPreviousConstituentInType,
                    ***REMOVED***, node);
                    return uniqueConstituents;
                ***REMOVED***
                cachedTypeMap.set(constituentNodeType, constituentNode);
                return [...uniqueConstituents, constituentNode];
            ***REMOVED***, []);
        ***REMOVED***
        function reportDuplicate(duplicateConstituent, parentNode) {
            const sourceCode = context.getSourceCode();
            const beforeTokens = sourceCode.getTokensBefore(duplicateConstituent.duplicated, { filter: token => token.value === '|' || token.value === '&' ***REMOVED***);
            const beforeUnionOrIntersectionToken = beforeTokens[beforeTokens.length - 1];
            const bracketBeforeTokens = sourceCode.getTokensBetween(beforeUnionOrIntersectionToken, duplicateConstituent.duplicated);
            const bracketAfterTokens = sourceCode.getTokensAfter(duplicateConstituent.duplicated, { count: bracketBeforeTokens.length ***REMOVED***);
            const reportLocation = {
                start: duplicateConstituent.duplicated.loc.start,
                end: bracketAfterTokens.length > 0
                    ? bracketAfterTokens[bracketAfterTokens.length - 1].loc.end
                    : duplicateConstituent.duplicated.loc.end,
            ***REMOVED***;
            context.report({
                data: {
                    type: parentNode.type === utils_1.AST_NODE_TYPES.TSIntersectionType
                        ? 'Intersection'
                        : 'Union',
                    previous: sourceCode.getText(duplicateConstituent.duplicatePrevious),
                ***REMOVED***,
                messageId: 'duplicate',
                node: duplicateConstituent.duplicated,
                loc: reportLocation,
                fix: fixer => {
                    return [
                        beforeUnionOrIntersectionToken,
                        ...bracketBeforeTokens,
                        duplicateConstituent.duplicated,
                        ...bracketAfterTokens,
                    ].map(token => fixer.remove(token));
                ***REMOVED***,
            ***REMOVED***);
        ***REMOVED***
        return Object.assign(Object.assign({***REMOVED***, (!ignoreIntersections && {
            TSIntersectionType: checkDuplicate,
        ***REMOVED***)), (!ignoreUnions && {
            TSUnionType: checkDuplicate,
        ***REMOVED***));
    ***REMOVED***,
***REMOVED***);
