// @typescript-eslint/eslint-plugin/dist/rules/no-extraneous-class.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; ***REMOVED*** ***REMOVED***;
    ***REMOVED***
    Object.defineProperty(o, k2, desc);
***REMOVED***) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
***REMOVED***));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v ***REMOVED***);
***REMOVED***) : function(o, v) {
    o["default"] = v;
***REMOVED***);
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {***REMOVED***;
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const utils_1 = require("@typescript-eslint/utils");
const util = __importStar(require("../util"));
exports.default = util.createRule({
    name: 'no-extraneous-class',
    meta: {
        type: 'suggestion',
        docs: {
            description: 'Disallow classes used as namespaces',
            recommended: 'strict',
        ***REMOVED***,
        schema: [
            {
                type: 'object',
                additionalProperties: false,
                properties: {
                    allowConstructorOnly: {
                        description: 'Whether to allow extraneous classes that contain only a constructor.',
                        type: 'boolean',
                    ***REMOVED***,
                    allowEmpty: {
                        description: 'Whether to allow extraneous classes that have no body (i.e. are empty).',
                        type: 'boolean',
                    ***REMOVED***,
                    allowStaticOnly: {
                        description: 'Whether to allow extraneous classes that only contain static members.',
                        type: 'boolean',
                    ***REMOVED***,
                    allowWithDecorator: {
                        description: 'Whether to allow extraneous classes that include a decorator.',
                        type: 'boolean',
                    ***REMOVED***,
                ***REMOVED***,
            ***REMOVED***,
        ],
        messages: {
            empty: 'Unexpected empty class.',
            onlyStatic: 'Unexpected class with only static properties.',
            onlyConstructor: 'Unexpected class with only a constructor.',
        ***REMOVED***,
    ***REMOVED***,
    defaultOptions: [
        {
            allowConstructorOnly: false,
            allowEmpty: false,
            allowStaticOnly: false,
            allowWithDecorator: false,
        ***REMOVED***,
    ],
    create(context, [{ allowConstructorOnly, allowEmpty, allowStaticOnly, allowWithDecorator ***REMOVED***]) {
        const isAllowWithDecorator = (node) => {
            return !!(allowWithDecorator &&
                node &&
                node.decorators &&
                node.decorators.length);
        ***REMOVED***;
        return {
            ClassBody(node) {
                const parent = node.parent;
                if (!parent || parent.superClass || isAllowWithDecorator(parent)) {
                    return;
                ***REMOVED***
                const reportNode = 'id' in parent && parent.id ? parent.id : parent;
                if (node.body.length === 0) {
                    if (allowEmpty) {
                        return;
                    ***REMOVED***
                    context.report({
                        node: reportNode,
                        messageId: 'empty',
                    ***REMOVED***);
                    return;
                ***REMOVED***
                let onlyStatic = true;
                let onlyConstructor = true;
                for (const prop of node.body) {
                    if ('kind' in prop && prop.kind === 'constructor') {
                        if (prop.value.params.some(param => param.type === utils_1.AST_NODE_TYPES.TSParameterProperty)) {
                            onlyConstructor = false;
                            onlyStatic = false;
                        ***REMOVED***
                    ***REMOVED***
                    else {
                        onlyConstructor = false;
                        if ('static' in prop && !prop.static) {
                            onlyStatic = false;
                        ***REMOVED***
                    ***REMOVED***
                    if (!(onlyStatic || onlyConstructor)) {
                        break;
                    ***REMOVED***
                ***REMOVED***
                if (onlyConstructor) {
                    if (!allowConstructorOnly) {
                        context.report({
                            node: reportNode,
                            messageId: 'onlyConstructor',
                        ***REMOVED***);
                    ***REMOVED***
                    return;
                ***REMOVED***
                if (onlyStatic && !allowStaticOnly) {
                    context.report({
                        node: reportNode,
                        messageId: 'onlyStatic',
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***;
    ***REMOVED***,
***REMOVED***);
