// @typescript-eslint/eslint-plugin/dist/rules/no-invalid-this.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const utils_1 = require("@typescript-eslint/utils");
const util_1 = require("../util");
const getESLintCoreRule_1 = require("../util/getESLintCoreRule");
const baseRule = (0, getESLintCoreRule_1.getESLintCoreRule)('no-invalid-this');
exports.default = (0, util_1.createRule)({
    name: 'no-invalid-this',
    meta: {
        type: 'suggestion',
        docs: {
            description: 'Disallow `this` keywords outside of classes or class-like objects',
            recommended: false,
            extendsBaseRule: true,
        ***REMOVED***,
        // TODO: this rule has only had messages since v7.0 - remove this when we remove support for v6
        messages: (_a = baseRule.meta.messages) !== null && _a !== void 0 ? _a : {
            unexpectedThis: "Unexpected 'this'.",
        ***REMOVED***,
        hasSuggestions: baseRule.meta.hasSuggestions,
        schema: baseRule.meta.schema,
    ***REMOVED***,
    defaultOptions: [{ capIsConstructor: true ***REMOVED***],
    create(context) {
        const rules = baseRule.create(context);
        /**
         * Since function definitions can be nested we use a stack storing if "this" is valid in the current context.
         *
         * Example:
         *
         * function a(this: number) { // valid "this"
         *     function b() {
         *         console.log(this); // invalid "this"
         *     ***REMOVED***
         * ***REMOVED***
         *
         * When parsing the function declaration of "a" the stack will be: [true]
         * When parsing the function declaration of "b" the stack will be: [true, false]
         */
        const thisIsValidStack = [];
        return Object.assign(Object.assign({***REMOVED***, rules), { PropertyDefinition() {
                thisIsValidStack.push(true);
            ***REMOVED***,
            'PropertyDefinition:exit'() {
                thisIsValidStack.pop();
            ***REMOVED***,
            FunctionDeclaration(node) {
                var _a;
                thisIsValidStack.push(node.params.some(param => param.type === utils_1.AST_NODE_TYPES.Identifier && param.name === 'this'));
                // baseRule's work
                (_a = rules.FunctionDeclaration) === null || _a === void 0 ? void 0 : _a.call(rules, node);
            ***REMOVED***,
            'FunctionDeclaration:exit'(node) {
                var _a;
                thisIsValidStack.pop();
                // baseRule's work
                (_a = rules['FunctionDeclaration:exit']) === null || _a === void 0 ? void 0 : _a.call(rules, node);
            ***REMOVED***,
            FunctionExpression(node) {
                var _a;
                thisIsValidStack.push(node.params.some(param => param.type === utils_1.AST_NODE_TYPES.Identifier && param.name === 'this'));
                // baseRule's work
                (_a = rules.FunctionExpression) === null || _a === void 0 ? void 0 : _a.call(rules, node);
            ***REMOVED***,
            'FunctionExpression:exit'(node) {
                var _a;
                thisIsValidStack.pop();
                // baseRule's work
                (_a = rules['FunctionExpression:exit']) === null || _a === void 0 ? void 0 : _a.call(rules, node);
            ***REMOVED***,
            ThisExpression(node) {
                const thisIsValidHere = thisIsValidStack[thisIsValidStack.length - 1];
                if (thisIsValidHere) {
                    return;
                ***REMOVED***
                // baseRule's work
                rules.ThisExpression(node);
            ***REMOVED*** ***REMOVED***);
    ***REMOVED***,
***REMOVED***);
