// @typescript-eslint/eslint-plugin/dist/rules/no-parameter-properties.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; ***REMOVED*** ***REMOVED***;
    ***REMOVED***
    Object.defineProperty(o, k2, desc);
***REMOVED***) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
***REMOVED***));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v ***REMOVED***);
***REMOVED***) : function(o, v) {
    o["default"] = v;
***REMOVED***);
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {***REMOVED***;
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const utils_1 = require("@typescript-eslint/utils");
const util = __importStar(require("../util"));
exports.default = util.createRule({
    name: 'no-parameter-properties',
    meta: {
        deprecated: true,
        replacedBy: ['@typescript-eslint/parameter-properties'],
        type: 'problem',
        docs: {
            description: 'Disallow the use of parameter properties in class constructors',
            // too opinionated to be recommended
            recommended: false,
        ***REMOVED***,
        messages: {
            noParamProp: 'Property {{parameter***REMOVED******REMOVED*** cannot be declared in the constructor.',
        ***REMOVED***,
        schema: [
            {
                type: 'object',
                properties: {
                    allows: {
                        type: 'array',
                        items: {
                            enum: [
                                'readonly',
                                'private',
                                'protected',
                                'public',
                                'private readonly',
                                'protected readonly',
                                'public readonly',
                            ],
                        ***REMOVED***,
                        minItems: 1,
                    ***REMOVED***,
                ***REMOVED***,
                additionalProperties: false,
            ***REMOVED***,
        ],
    ***REMOVED***,
    defaultOptions: [
        {
            allows: [],
        ***REMOVED***,
    ],
    create(context, [{ allows ***REMOVED***]) {
        /**
         * Gets the modifiers of `node`.
         * @param node the node to be inspected.
         */
        function getModifiers(node) {
            const modifiers = [];
            if (node.accessibility) {
                modifiers.push(node.accessibility);
            ***REMOVED***
            if (node.readonly) {
                modifiers.push('readonly');
            ***REMOVED***
            return modifiers.filter(Boolean).join(' ');
        ***REMOVED***
        return {
            TSParameterProperty(node) {
                const modifiers = getModifiers(node);
                if (!allows.includes(modifiers)) {
                    // HAS to be an identifier or assignment or TSC will throw
                    if (node.parameter.type !== utils_1.AST_NODE_TYPES.Identifier &&
                        node.parameter.type !== utils_1.AST_NODE_TYPES.AssignmentPattern) {
                        return;
                    ***REMOVED***
                    const name = node.parameter.type === utils_1.AST_NODE_TYPES.Identifier
                        ? node.parameter.name
                        : // has to be an Identifier or TSC will throw an error
                            node.parameter.left.name;
                    context.report({
                        node,
                        messageId: 'noParamProp',
                        data: {
                            parameter: name,
                        ***REMOVED***,
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***;
    ***REMOVED***,
***REMOVED***);
