// @typescript-eslint/eslint-plugin/dist/rules/no-unnecessary-qualifier.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; ***REMOVED*** ***REMOVED***;
    ***REMOVED***
    Object.defineProperty(o, k2, desc);
***REMOVED***) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
***REMOVED***));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v ***REMOVED***);
***REMOVED***) : function(o, v) {
    o["default"] = v;
***REMOVED***);
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {***REMOVED***;
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const utils_1 = require("@typescript-eslint/utils");
const tsutils = __importStar(require("tsutils"));
const ts = __importStar(require("typescript"));
const util = __importStar(require("../util"));
exports.default = util.createRule({
    name: 'no-unnecessary-qualifier',
    meta: {
        docs: {
            description: 'Disallow unnecessary namespace qualifiers',
            recommended: false,
            requiresTypeChecking: true,
        ***REMOVED***,
        fixable: 'code',
        messages: {
            unnecessaryQualifier: "Qualifier is unnecessary since '{{ name ***REMOVED******REMOVED***' is in scope.",
        ***REMOVED***,
        schema: [],
        type: 'suggestion',
    ***REMOVED***,
    defaultOptions: [],
    create(context) {
        const namespacesInScope = [];
        let currentFailedNamespaceExpression = null;
        const parserServices = util.getParserServices(context);
        const esTreeNodeToTSNodeMap = parserServices.esTreeNodeToTSNodeMap;
        const program = parserServices.program;
        const checker = program.getTypeChecker();
        const sourceCode = context.getSourceCode();
        function tryGetAliasedSymbol(symbol, checker) {
            return tsutils.isSymbolFlagSet(symbol, ts.SymbolFlags.Alias)
                ? checker.getAliasedSymbol(symbol)
                : null;
        ***REMOVED***
        function symbolIsNamespaceInScope(symbol) {
            var _a;
            const symbolDeclarations = (_a = symbol.getDeclarations()) !== null && _a !== void 0 ? _a : [];
            if (symbolDeclarations.some(decl => namespacesInScope.some(ns => ns === decl))) {
                return true;
            ***REMOVED***
            const alias = tryGetAliasedSymbol(symbol, checker);
            return alias != null && symbolIsNamespaceInScope(alias);
        ***REMOVED***
        function getSymbolInScope(node, flags, name) {
            // TODO:PERF `getSymbolsInScope` gets a long list. Is there a better way?
            const scope = checker.getSymbolsInScope(node, flags);
            return scope.find(scopeSymbol => scopeSymbol.name === name);
        ***REMOVED***
        function symbolsAreEqual(accessed, inScope) {
            return accessed === checker.getExportSymbolOfSymbol(inScope);
        ***REMOVED***
        function qualifierIsUnnecessary(qualifier, name) {
            const tsQualifier = esTreeNodeToTSNodeMap.get(qualifier);
            const tsName = esTreeNodeToTSNodeMap.get(name);
            const namespaceSymbol = checker.getSymbolAtLocation(tsQualifier);
            if (namespaceSymbol === undefined ||
                !symbolIsNamespaceInScope(namespaceSymbol)) {
                return false;
            ***REMOVED***
            const accessedSymbol = checker.getSymbolAtLocation(tsName);
            if (accessedSymbol === undefined) {
                return false;
            ***REMOVED***
            // If the symbol in scope is different, the qualifier is necessary.
            const fromScope = getSymbolInScope(tsQualifier, accessedSymbol.flags, sourceCode.getText(name));
            return (fromScope === undefined || symbolsAreEqual(accessedSymbol, fromScope));
        ***REMOVED***
        function visitNamespaceAccess(node, qualifier, name) {
            // Only look for nested qualifier errors if we didn't already fail on the outer qualifier.
            if (!currentFailedNamespaceExpression &&
                qualifierIsUnnecessary(qualifier, name)) {
                currentFailedNamespaceExpression = node;
                context.report({
                    node: qualifier,
                    messageId: 'unnecessaryQualifier',
                    data: {
                        name: sourceCode.getText(name),
                    ***REMOVED***,
                    fix(fixer) {
                        return fixer.removeRange([qualifier.range[0], name.range[0]]);
                    ***REMOVED***,
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***
        function enterDeclaration(node) {
            namespacesInScope.push(esTreeNodeToTSNodeMap.get(node));
        ***REMOVED***
        function exitDeclaration() {
            namespacesInScope.pop();
        ***REMOVED***
        function resetCurrentNamespaceExpression(node) {
            if (node === currentFailedNamespaceExpression) {
                currentFailedNamespaceExpression = null;
            ***REMOVED***
        ***REMOVED***
        function isPropertyAccessExpression(node) {
            return node.type === utils_1.AST_NODE_TYPES.MemberExpression && !node.computed;
        ***REMOVED***
        function isEntityNameExpression(node) {
            return (node.type === utils_1.AST_NODE_TYPES.Identifier ||
                (isPropertyAccessExpression(node) &&
                    isEntityNameExpression(node.object)));
        ***REMOVED***
        return {
            TSModuleDeclaration: enterDeclaration,
            TSEnumDeclaration: enterDeclaration,
            'ExportNamedDeclaration[declaration.type="TSModuleDeclaration"]': enterDeclaration,
            'ExportNamedDeclaration[declaration.type="TSEnumDeclaration"]': enterDeclaration,
            'TSModuleDeclaration:exit': exitDeclaration,
            'TSEnumDeclaration:exit': exitDeclaration,
            'ExportNamedDeclaration[declaration.type="TSModuleDeclaration"]:exit': exitDeclaration,
            'ExportNamedDeclaration[declaration.type="TSEnumDeclaration"]:exit': exitDeclaration,
            TSQualifiedName(node) {
                visitNamespaceAccess(node, node.left, node.right);
            ***REMOVED***,
            'MemberExpression[computed=false]': function (node) {
                const property = node.property;
                if (isEntityNameExpression(node.object)) {
                    visitNamespaceAccess(node, node.object, property);
                ***REMOVED***
            ***REMOVED***,
            'TSQualifiedName:exit': resetCurrentNamespaceExpression,
            'MemberExpression:exit': resetCurrentNamespaceExpression,
        ***REMOVED***;
    ***REMOVED***,
***REMOVED***);
