// @typescript-eslint/eslint-plugin/dist/rules/no-unused-vars.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; ***REMOVED*** ***REMOVED***;
    ***REMOVED***
    Object.defineProperty(o, k2, desc);
***REMOVED***) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
***REMOVED***));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v ***REMOVED***);
***REMOVED***) : function(o, v) {
    o["default"] = v;
***REMOVED***);
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {***REMOVED***;
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const scope_manager_1 = require("@typescript-eslint/scope-manager");
const utils_1 = require("@typescript-eslint/utils");
const util = __importStar(require("../util"));
exports.default = util.createRule({
    name: 'no-unused-vars',
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallow unused variables',
            recommended: 'warn',
            extendsBaseRule: true,
        ***REMOVED***,
        schema: [
            {
                oneOf: [
                    {
                        enum: ['all', 'local'],
                    ***REMOVED***,
                    {
                        type: 'object',
                        properties: {
                            vars: {
                                enum: ['all', 'local'],
                            ***REMOVED***,
                            varsIgnorePattern: {
                                type: 'string',
                            ***REMOVED***,
                            args: {
                                enum: ['all', 'after-used', 'none'],
                            ***REMOVED***,
                            ignoreRestSiblings: {
                                type: 'boolean',
                            ***REMOVED***,
                            argsIgnorePattern: {
                                type: 'string',
                            ***REMOVED***,
                            caughtErrors: {
                                enum: ['all', 'none'],
                            ***REMOVED***,
                            caughtErrorsIgnorePattern: {
                                type: 'string',
                            ***REMOVED***,
                            destructuredArrayIgnorePattern: {
                                type: 'string',
                            ***REMOVED***,
                        ***REMOVED***,
                        additionalProperties: false,
                    ***REMOVED***,
                ],
            ***REMOVED***,
        ],
        messages: {
            unusedVar: "'{{varName***REMOVED******REMOVED***' is {{action***REMOVED******REMOVED*** but never used{{additional***REMOVED******REMOVED***.",
        ***REMOVED***,
    ***REMOVED***,
    defaultOptions: [{***REMOVED***],
    create(context, [firstOption]) {
        const filename = context.getFilename();
        const sourceCode = context.getSourceCode();
        const MODULE_DECL_CACHE = new Map();
        const options = (() => {
            var _a, _b, _c, _d;
            const options = {
                vars: 'all',
                args: 'after-used',
                ignoreRestSiblings: false,
                caughtErrors: 'none',
            ***REMOVED***;
            if (firstOption) {
                if (typeof firstOption === 'string') {
                    options.vars = firstOption;
                ***REMOVED***
                else {
                    options.vars = (_a = firstOption.vars) !== null && _a !== void 0 ? _a : options.vars;
                    options.args = (_b = firstOption.args) !== null && _b !== void 0 ? _b : options.args;
                    options.ignoreRestSiblings =
                        (_c = firstOption.ignoreRestSiblings) !== null && _c !== void 0 ? _c : options.ignoreRestSiblings;
                    options.caughtErrors =
                        (_d = firstOption.caughtErrors) !== null && _d !== void 0 ? _d : options.caughtErrors;
                    if (firstOption.varsIgnorePattern) {
                        options.varsIgnorePattern = new RegExp(firstOption.varsIgnorePattern, 'u');
                    ***REMOVED***
                    if (firstOption.argsIgnorePattern) {
                        options.argsIgnorePattern = new RegExp(firstOption.argsIgnorePattern, 'u');
                    ***REMOVED***
                    if (firstOption.caughtErrorsIgnorePattern) {
                        options.caughtErrorsIgnorePattern = new RegExp(firstOption.caughtErrorsIgnorePattern, 'u');
                    ***REMOVED***
                    if (firstOption.destructuredArrayIgnorePattern) {
                        options.destructuredArrayIgnorePattern = new RegExp(firstOption.destructuredArrayIgnorePattern, 'u');
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
            return options;
        ***REMOVED***)();
        function collectUnusedVariables() {
            var _a, _b, _c, _d, _e;
            /**
             * Checks whether a node is a sibling of the rest property or not.
             * @param {ASTNode***REMOVED*** node a node to check
             * @returns {boolean***REMOVED*** True if the node is a sibling of the rest property, otherwise false.
             */
            function hasRestSibling(node) {
                var _a;
                return (node.type === utils_1.AST_NODE_TYPES.Property &&
                    ((_a = node.parent) === null || _a === void 0 ? void 0 : _a.type) === utils_1.AST_NODE_TYPES.ObjectPattern &&
                    node.parent.properties[node.parent.properties.length - 1].type ===
                        utils_1.AST_NODE_TYPES.RestElement);
            ***REMOVED***
            /**
             * Determines if a variable has a sibling rest property
             * @param variable eslint-scope variable object.
             * @returns True if the variable is exported, false if not.
             */
            function hasRestSpreadSibling(variable) {
                if (options.ignoreRestSiblings) {
                    const hasRestSiblingDefinition = variable.defs.some(def => hasRestSibling(def.name.parent));
                    const hasRestSiblingReference = variable.references.some(ref => hasRestSibling(ref.identifier.parent));
                    return hasRestSiblingDefinition || hasRestSiblingReference;
                ***REMOVED***
                return false;
            ***REMOVED***
            /**
             * Checks whether the given variable is after the last used parameter.
             * @param variable The variable to check.
             * @returns `true` if the variable is defined after the last used parameter.
             */
            function isAfterLastUsedArg(variable) {
                const def = variable.defs[0];
                const params = context.getDeclaredVariables(def.node);
                const posteriorParams = params.slice(params.indexOf(variable) + 1);
                // If any used parameters occur after this parameter, do not report.
                return !posteriorParams.some(v => v.references.length > 0 || v.eslintUsed);
            ***REMOVED***
            const unusedVariablesOriginal = util.collectUnusedVariables(context);
            const unusedVariablesReturn = [];
            for (const variable of unusedVariablesOriginal) {
                // explicit global variables don't have definitions.
                if (variable.defs.length === 0) {
                    unusedVariablesReturn.push(variable);
                    continue;
                ***REMOVED***
                const def = variable.defs[0];
                if (variable.scope.type === utils_1.TSESLint.Scope.ScopeType.global &&
                    options.vars === 'local') {
                    // skip variables in the global scope if configured to
                    continue;
                ***REMOVED***
                const refUsedInArrayPatterns = variable.references.some(ref => { var _a; return ((_a = ref.identifier.parent) === null || _a === void 0 ? void 0 : _a.type) === utils_1.AST_NODE_TYPES.ArrayPattern; ***REMOVED***);
                // skip elements of array destructuring patterns
                if ((((_a = def.name.parent) === null || _a === void 0 ? void 0 : _a.type) === utils_1.AST_NODE_TYPES.ArrayPattern ||
                    refUsedInArrayPatterns) &&
                    'name' in def.name &&
                    ((_b = options.destructuredArrayIgnorePattern) === null || _b === void 0 ? void 0 : _b.test(def.name.name))) {
                    continue;
                ***REMOVED***
                // skip catch variables
                if (def.type === utils_1.TSESLint.Scope.DefinitionType.CatchClause) {
                    if (options.caughtErrors === 'none') {
                        continue;
                    ***REMOVED***
                    // skip ignored parameters
                    if ('name' in def.name &&
                        ((_c = options.caughtErrorsIgnorePattern) === null || _c === void 0 ? void 0 : _c.test(def.name.name))) {
                        continue;
                    ***REMOVED***
                ***REMOVED***
                if (def.type === utils_1.TSESLint.Scope.DefinitionType.Parameter) {
                    // if "args" option is "none", skip any parameter
                    if (options.args === 'none') {
                        continue;
                    ***REMOVED***
                    // skip ignored parameters
                    if ('name' in def.name &&
                        ((_d = options.argsIgnorePattern) === null || _d === void 0 ? void 0 : _d.test(def.name.name))) {
                        continue;
                    ***REMOVED***
                    // if "args" option is "after-used", skip used variables
                    if (options.args === 'after-used' &&
                        util.isFunction(def.name.parent) &&
                        !isAfterLastUsedArg(variable)) {
                        continue;
                    ***REMOVED***
                ***REMOVED***
                else {
                    // skip ignored variables
                    if ('name' in def.name &&
                        ((_e = options.varsIgnorePattern) === null || _e === void 0 ? void 0 : _e.test(def.name.name))) {
                        continue;
                    ***REMOVED***
                ***REMOVED***
                if (hasRestSpreadSibling(variable)) {
                    continue;
                ***REMOVED***
                // in case another rule has run and used the collectUnusedVariables,
                // we want to ensure our selectors that marked variables as used are respected
                if (variable.eslintUsed) {
                    continue;
                ***REMOVED***
                unusedVariablesReturn.push(variable);
            ***REMOVED***
            return unusedVariablesReturn;
        ***REMOVED***
        return {
            // declaration file handling
            [ambientDeclarationSelector(utils_1.AST_NODE_TYPES.Program, true)](node) {
                if (!util.isDefinitionFile(filename)) {
                    return;
                ***REMOVED***
                markDeclarationChildAsUsed(node);
            ***REMOVED***,
            // module declaration in module declaration should not report unused vars error
            // this is workaround as this change should be done in better way
            'TSModuleDeclaration > TSModuleDeclaration'(node) {
                if (node.id.type === utils_1.AST_NODE_TYPES.Identifier) {
                    let scope = context.getScope();
                    if (scope.upper) {
                        scope = scope.upper;
                    ***REMOVED***
                    const superVar = scope.set.get(node.id.name);
                    if (superVar) {
                        superVar.eslintUsed = true;
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***,
            // children of a namespace that is a child of a declared namespace are auto-exported
            [ambientDeclarationSelector('TSModuleDeclaration[declare = true] > TSModuleBlock TSModuleDeclaration > TSModuleBlock', false)](node) {
                markDeclarationChildAsUsed(node);
            ***REMOVED***,
            // declared namespace handling
            [ambientDeclarationSelector('TSModuleDeclaration[declare = true] > TSModuleBlock', false)](node) {
                var _a;
                const moduleDecl = util.nullThrows((_a = node.parent) === null || _a === void 0 ? void 0 : _a.parent, util.NullThrowsReasons.MissingParent);
                // declared ambient modules with an `export =` statement will only export that one thing
                // all other statements are not automatically exported in this case
                if (moduleDecl.id.type === utils_1.AST_NODE_TYPES.Literal &&
                    checkModuleDeclForExportEquals(moduleDecl)) {
                    return;
                ***REMOVED***
                markDeclarationChildAsUsed(node);
            ***REMOVED***,
            // collect
            'Program:exit'(programNode) {
                /**
                 * Generates the message data about the variable being defined and unused,
                 * including the ignore pattern if configured.
                 * @param unusedVar eslint-scope variable object.
                 * @returns The message data to be used with this unused variable.
                 */
                function getDefinedMessageData(unusedVar) {
                    var _a;
                    const defType = (_a = unusedVar === null || unusedVar === void 0 ? void 0 : unusedVar.defs[0]) === null || _a === void 0 ? void 0 : _a.type;
                    let type;
                    let pattern;
                    if (defType === utils_1.TSESLint.Scope.DefinitionType.CatchClause &&
                        options.caughtErrorsIgnorePattern) {
                        type = 'args';
                        pattern = options.caughtErrorsIgnorePattern.toString();
                    ***REMOVED***
                    else if (defType === utils_1.TSESLint.Scope.DefinitionType.Parameter &&
                        options.argsIgnorePattern) {
                        type = 'args';
                        pattern = options.argsIgnorePattern.toString();
                    ***REMOVED***
                    else if (defType !== utils_1.TSESLint.Scope.DefinitionType.Parameter &&
                        options.varsIgnorePattern) {
                        type = 'vars';
                        pattern = options.varsIgnorePattern.toString();
                    ***REMOVED***
                    const additional = type
                        ? `. Allowed unused ${type***REMOVED*** must match ${pattern***REMOVED***`
                        : '';
                    return {
                        varName: unusedVar.name,
                        action: 'defined',
                        additional,
                    ***REMOVED***;
                ***REMOVED***
                /**
                 * Generate the warning message about the variable being
                 * assigned and unused, including the ignore pattern if configured.
                 * @param unusedVar eslint-scope variable object.
                 * @returns The message data to be used with this unused variable.
                 */
                function getAssignedMessageData(unusedVar) {
                    var _a;
                    const def = unusedVar.defs[0];
                    let additional = '';
                    if (options.destructuredArrayIgnorePattern &&
                        ((_a = def === null || def === void 0 ? void 0 : def.name.parent) === null || _a === void 0 ? void 0 : _a.type) === utils_1.AST_NODE_TYPES.ArrayPattern) {
                        additional = `. Allowed unused elements of array destructuring patterns must match ${options.destructuredArrayIgnorePattern.toString()***REMOVED***`;
                    ***REMOVED***
                    else if (options.varsIgnorePattern) {
                        additional = `. Allowed unused vars must match ${options.varsIgnorePattern.toString()***REMOVED***`;
                    ***REMOVED***
                    return {
                        varName: unusedVar.name,
                        action: 'assigned a value',
                        additional,
                    ***REMOVED***;
                ***REMOVED***
                const unusedVars = collectUnusedVariables();
                for (const unusedVar of unusedVars) {
                    // Report the first declaration.
                    if (unusedVar.defs.length > 0) {
                        const writeReferences = unusedVar.references.filter(ref => ref.isWrite() &&
                            ref.from.variableScope === unusedVar.scope.variableScope);
                        context.report({
                            node: writeReferences.length
                                ? writeReferences[writeReferences.length - 1].identifier
                                : unusedVar.identifiers[0],
                            messageId: 'unusedVar',
                            data: unusedVar.references.some(ref => ref.isWrite())
                                ? getAssignedMessageData(unusedVar)
                                : getDefinedMessageData(unusedVar),
                        ***REMOVED***);
                        // If there are no regular declaration, report the first `/*globals*/` comment directive.
                    ***REMOVED***
                    else if ('eslintExplicitGlobalComments' in unusedVar &&
                        unusedVar.eslintExplicitGlobalComments) {
                        const directiveComment = unusedVar.eslintExplicitGlobalComments[0];
                        context.report({
                            node: programNode,
                            loc: util.getNameLocationInGlobalDirectiveComment(sourceCode, directiveComment, unusedVar.name),
                            messageId: 'unusedVar',
                            data: getDefinedMessageData(unusedVar),
                        ***REMOVED***);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***;
        function checkModuleDeclForExportEquals(node) {
            const cached = MODULE_DECL_CACHE.get(node);
            if (cached != null) {
                return cached;
            ***REMOVED***
            if (node.body && node.body.type === utils_1.AST_NODE_TYPES.TSModuleBlock) {
                for (const statement of node.body.body) {
                    if (statement.type === utils_1.AST_NODE_TYPES.TSExportAssignment) {
                        MODULE_DECL_CACHE.set(node, true);
                        return true;
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
            MODULE_DECL_CACHE.set(node, false);
            return false;
        ***REMOVED***
        function ambientDeclarationSelector(parent, childDeclare) {
            return [
                // Types are ambiently exported
                `${parent***REMOVED*** > :matches(${[
                    utils_1.AST_NODE_TYPES.TSInterfaceDeclaration,
                    utils_1.AST_NODE_TYPES.TSTypeAliasDeclaration,
                ].join(', ')***REMOVED***)`,
                // Value things are ambiently exported if they are "declare"d
                `${parent***REMOVED*** > :matches(${[
                    utils_1.AST_NODE_TYPES.ClassDeclaration,
                    utils_1.AST_NODE_TYPES.TSDeclareFunction,
                    utils_1.AST_NODE_TYPES.TSEnumDeclaration,
                    utils_1.AST_NODE_TYPES.TSModuleDeclaration,
                    utils_1.AST_NODE_TYPES.VariableDeclaration,
                ].join(', ')***REMOVED***)${childDeclare ? '[declare = true]' : ''***REMOVED***`,
            ].join(', ');
        ***REMOVED***
        function markDeclarationChildAsUsed(node) {
            var _a;
            const identifiers = [];
            switch (node.type) {
                case utils_1.AST_NODE_TYPES.TSInterfaceDeclaration:
                case utils_1.AST_NODE_TYPES.TSTypeAliasDeclaration:
                case utils_1.AST_NODE_TYPES.ClassDeclaration:
                case utils_1.AST_NODE_TYPES.FunctionDeclaration:
                case utils_1.AST_NODE_TYPES.TSDeclareFunction:
                case utils_1.AST_NODE_TYPES.TSEnumDeclaration:
                case utils_1.AST_NODE_TYPES.TSModuleDeclaration:
                    if (((_a = node.id) === null || _a === void 0 ? void 0 : _a.type) === utils_1.AST_NODE_TYPES.Identifier) {
                        identifiers.push(node.id);
                    ***REMOVED***
                    break;
                case utils_1.AST_NODE_TYPES.VariableDeclaration:
                    for (const declaration of node.declarations) {
                        visitPattern(declaration, pattern => {
                            identifiers.push(pattern);
                        ***REMOVED***);
                    ***REMOVED***
                    break;
            ***REMOVED***
            let scope = context.getScope();
            const shouldUseUpperScope = [
                utils_1.AST_NODE_TYPES.TSModuleDeclaration,
                utils_1.AST_NODE_TYPES.TSDeclareFunction,
            ].includes(node.type);
            if (scope.variableScope !== scope) {
                scope = scope.variableScope;
            ***REMOVED***
            else if (shouldUseUpperScope && scope.upper) {
                scope = scope.upper;
            ***REMOVED***
            for (const id of identifiers) {
                const superVar = scope.set.get(id.name);
                if (superVar) {
                    superVar.eslintUsed = true;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        function visitPattern(node, cb) {
            const visitor = new scope_manager_1.PatternVisitor({***REMOVED***, node, cb);
            visitor.visit(node);
        ***REMOVED***
    ***REMOVED***,
***REMOVED***);
/*

###### TODO ######

Edge cases that aren't currently handled due to laziness and them being super edgy edge cases


--- function params referenced in typeof type refs in the function declaration ---
--- NOTE - TS gets these cases wrong

function _foo(
  arg: number // arg should be unused
): typeof arg {
  return 1 as any;
***REMOVED***

function _bar(
  arg: number, // arg should be unused
  _arg2: typeof arg,
) {***REMOVED***


--- function names referenced in typeof type refs in the function declaration ---
--- NOTE - TS gets these cases right

function foo( // foo should be unused
): typeof foo {
    return 1 as any;
***REMOVED***

function bar( // bar should be unused
  _arg: typeof bar
) {***REMOVED***


--- if an interface is merged into a namespace  ---
--- NOTE - TS gets these cases wrong

namespace Test {
    interface Foo { // Foo should be unused here
        a: string;
    ***REMOVED***
    export namespace Foo {
       export type T = 'b';
    ***REMOVED***
***REMOVED***
type T = Test.Foo; // Error: Namespace 'Test' has no exported member 'Foo'.


namespace Test {
    export interface Foo {
        a: string;
    ***REMOVED***
    namespace Foo { // Foo should be unused here
       export type T = 'b';
    ***REMOVED***
***REMOVED***
type T = Test.Foo.T; // Error: Namespace 'Test' has no exported member 'Foo'.

*/
/*

###### TODO ######

We currently extend base `no-unused-vars` implementation because it's easier and lighter-weight.

Because of this, there are a few false-negatives which won't get caught.
We could fix these if we fork the base rule; but that's a lot of code (~650 lines) to add in.
I didn't want to do that just yet without some real-world issues, considering these are pretty rare edge-cases.

These cases are mishandled because the base rule assumes that each variable has one def, but type-value shadowing
creates a variable with two defs

--- type-only or value-only references to type/value shadowed variables ---
--- NOTE - TS gets these cases wrong

type T = 1;
const T = 2; // this T should be unused

type U = T; // this U should be unused
const U = 3;

const _V = U;


--- partially exported type/value shadowed variables ---
--- NOTE - TS gets these cases wrong

export interface Foo {***REMOVED***
const Foo = 1; // this Foo should be unused

interface Bar {***REMOVED*** // this Bar should be unused
export const Bar = 1;

*/
