// @typescript-eslint/eslint-plugin/dist/rules/require-array-sort-compare.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; ***REMOVED*** ***REMOVED***;
    ***REMOVED***
    Object.defineProperty(o, k2, desc);
***REMOVED***) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
***REMOVED***));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v ***REMOVED***);
***REMOVED***) : function(o, v) {
    o["default"] = v;
***REMOVED***);
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {***REMOVED***;
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const util = __importStar(require("../util"));
exports.default = util.createRule({
    name: 'require-array-sort-compare',
    defaultOptions: [
        {
            ignoreStringArrays: false,
        ***REMOVED***,
    ],
    meta: {
        type: 'problem',
        docs: {
            description: 'Require `Array#sort` calls to always provide a `compareFunction`',
            recommended: false,
            requiresTypeChecking: true,
        ***REMOVED***,
        messages: {
            requireCompare: "Require 'compare' argument.",
        ***REMOVED***,
        schema: [
            {
                type: 'object',
                properties: {
                    ignoreStringArrays: {
                        description: 'Whether to ignore arrays in which all elements are strings.',
                        type: 'boolean',
                    ***REMOVED***,
                ***REMOVED***,
            ***REMOVED***,
        ],
    ***REMOVED***,
    create(context, [options]) {
        const service = util.getParserServices(context);
        const checker = service.program.getTypeChecker();
        /**
         * Check if a given node is an array which all elements are string.
         * @param node
         */
        function isStringArrayNode(node) {
            const type = checker.getTypeAtLocation(service.esTreeNodeToTSNodeMap.get(node));
            if (checker.isArrayType(type) || checker.isTupleType(type)) {
                const typeArgs = util.getTypeArguments(type, checker);
                return typeArgs.every(arg => util.getTypeName(checker, arg) === 'string');
            ***REMOVED***
            return false;
        ***REMOVED***
        return {
            "CallExpression[arguments.length=0] > MemberExpression[property.name='sort'][computed=false]"(callee) {
                const tsNode = service.esTreeNodeToTSNodeMap.get(callee.object);
                const calleeObjType = util.getConstrainedTypeAtLocation(checker, tsNode);
                if (options.ignoreStringArrays && isStringArrayNode(callee.object)) {
                    return;
                ***REMOVED***
                if (util.isTypeArrayTypeOrUnionOfArrayTypes(calleeObjType, checker)) {
                    context.report({ node: callee.parent, messageId: 'requireCompare' ***REMOVED***);
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***;
    ***REMOVED***,
***REMOVED***);
