// @typescript-eslint/eslint-plugin/dist/rules/restrict-plus-operands.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; ***REMOVED*** ***REMOVED***;
    ***REMOVED***
    Object.defineProperty(o, k2, desc);
***REMOVED***) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
***REMOVED***));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v ***REMOVED***);
***REMOVED***) : function(o, v) {
    o["default"] = v;
***REMOVED***);
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {***REMOVED***;
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const tsutils = __importStar(require("tsutils"));
const ts = __importStar(require("typescript"));
const util = __importStar(require("../util"));
exports.default = util.createRule({
    name: 'restrict-plus-operands',
    meta: {
        type: 'problem',
        docs: {
            description: 'Require both operands of addition to be the same type and be `bigint`, `number`, or `string`',
            recommended: 'error',
            requiresTypeChecking: true,
        ***REMOVED***,
        messages: {
            bigintAndNumber: "Numeric '+' operations must either be both bigints or both numbers. Got `{{left***REMOVED******REMOVED***` + `{{right***REMOVED******REMOVED***`.",
            invalid: "Invalid operand for a '+' operation. Operands must each be a number or {{stringLike***REMOVED******REMOVED***. Got `{{type***REMOVED******REMOVED***`.",
            mismatched: "Operands of '+' operations must be a number or {{stringLike***REMOVED******REMOVED***. Got `{{left***REMOVED******REMOVED***` + `{{right***REMOVED******REMOVED***`.",
        ***REMOVED***,
        schema: [
            {
                type: 'object',
                additionalProperties: false,
                properties: {
                    allowAny: {
                        description: 'Whether to allow `any` typed values.',
                        type: 'boolean',
                    ***REMOVED***,
                    allowBoolean: {
                        description: 'Whether to allow `boolean` typed values.',
                        type: 'boolean',
                    ***REMOVED***,
                    allowNullish: {
                        description: 'Whether to allow potentially `null` or `undefined` typed values.',
                        type: 'boolean',
                    ***REMOVED***,
                    allowNumberAndString: {
                        description: 'Whether to allow `bigint`/`number` typed values and `string` typed values to be added together.',
                        type: 'boolean',
                    ***REMOVED***,
                    allowRegExp: {
                        description: 'Whether to allow `regexp` typed values.',
                        type: 'boolean',
                    ***REMOVED***,
                    checkCompoundAssignments: {
                        description: 'Whether to check compound assignments such as `+=`.',
                        type: 'boolean',
                    ***REMOVED***,
                ***REMOVED***,
            ***REMOVED***,
        ],
    ***REMOVED***,
    defaultOptions: [
        {
            checkCompoundAssignments: false,
        ***REMOVED***,
    ],
    create(context, [{ checkCompoundAssignments, allowAny, allowBoolean, allowNullish, allowNumberAndString, allowRegExp, ***REMOVED***,]) {
        const service = util.getParserServices(context);
        const typeChecker = service.program.getTypeChecker();
        const stringLikes = [
            allowAny && '`any`',
            allowBoolean && '`boolean`',
            allowNullish && '`null`',
            allowRegExp && '`RegExp`',
            allowNullish && '`undefined`',
        ].filter((value) => typeof value === 'string');
        const stringLike = stringLikes.length
            ? stringLikes.length === 1
                ? `string, allowing a string + ${stringLikes[0]***REMOVED***`
                : `string, allowing a string + any of: ${stringLikes.join(', ')***REMOVED***`
            : 'string';
        function getTypeConstrained(node) {
            return typeChecker.getBaseTypeOfLiteralType(util.getConstrainedTypeAtLocation(typeChecker, service.esTreeNodeToTSNodeMap.get(node)));
        ***REMOVED***
        function checkPlusOperands(node) {
            const leftType = getTypeConstrained(node.left);
            const rightType = getTypeConstrained(node.right);
            if (leftType === rightType &&
                tsutils.isTypeFlagSet(leftType, ts.TypeFlags.BigIntLike |
                    ts.TypeFlags.NumberLike |
                    ts.TypeFlags.StringLike)) {
                return;
            ***REMOVED***
            let hadIndividualComplaint = false;
            for (const [baseNode, baseType, otherType] of [
                [node.left, leftType, rightType],
                [node.right, rightType, leftType],
            ]) {
                if (isTypeFlagSetInUnion(baseType, ts.TypeFlags.ESSymbolLike |
                    ts.TypeFlags.Never |
                    ts.TypeFlags.Unknown) ||
                    (!allowAny && isTypeFlagSetInUnion(baseType, ts.TypeFlags.Any)) ||
                    (!allowBoolean &&
                        isTypeFlagSetInUnion(baseType, ts.TypeFlags.BooleanLike)) ||
                    (!allowNullish &&
                        util.isTypeFlagSet(baseType, ts.TypeFlags.Null | ts.TypeFlags.Undefined))) {
                    context.report({
                        data: {
                            stringLike,
                            type: typeChecker.typeToString(baseType),
                        ***REMOVED***,
                        messageId: 'invalid',
                        node: baseNode,
                    ***REMOVED***);
                    hadIndividualComplaint = true;
                    continue;
                ***REMOVED***
                // RegExps also contain ts.TypeFlags.Any & ts.TypeFlags.Object
                for (const subBaseType of tsutils.unionTypeParts(baseType)) {
                    const typeName = util.getTypeName(typeChecker, subBaseType);
                    if (typeName === 'RegExp'
                        ? !allowRegExp ||
                            tsutils.isTypeFlagSet(otherType, ts.TypeFlags.NumberLike)
                        : (!allowAny && util.isTypeAnyType(subBaseType)) ||
                            isDeeplyObjectType(subBaseType)) {
                        context.report({
                            data: {
                                stringLike,
                                type: typeChecker.typeToString(subBaseType),
                            ***REMOVED***,
                            messageId: 'invalid',
                            node: baseNode,
                        ***REMOVED***);
                        hadIndividualComplaint = true;
                        continue;
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
            if (hadIndividualComplaint) {
                return;
            ***REMOVED***
            for (const [baseType, otherType] of [
                [leftType, rightType],
                [rightType, leftType],
            ]) {
                if (!allowNumberAndString &&
                    isTypeFlagSetInUnion(baseType, ts.TypeFlags.StringLike) &&
                    isTypeFlagSetInUnion(otherType, ts.TypeFlags.NumberLike)) {
                    return context.report({
                        data: {
                            stringLike,
                            left: typeChecker.typeToString(leftType),
                            right: typeChecker.typeToString(rightType),
                        ***REMOVED***,
                        messageId: 'mismatched',
                        node,
                    ***REMOVED***);
                ***REMOVED***
                if (isTypeFlagSetInUnion(baseType, ts.TypeFlags.NumberLike) &&
                    isTypeFlagSetInUnion(otherType, ts.TypeFlags.BigIntLike)) {
                    return context.report({
                        data: {
                            left: typeChecker.typeToString(leftType),
                            right: typeChecker.typeToString(rightType),
                        ***REMOVED***,
                        messageId: 'bigintAndNumber',
                        node,
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        return Object.assign({ "BinaryExpression[operator='+']": checkPlusOperands ***REMOVED***, (checkCompoundAssignments && {
            "AssignmentExpression[operator='+=']"(node) {
                checkPlusOperands(node);
            ***REMOVED***,
        ***REMOVED***));
    ***REMOVED***,
***REMOVED***);
function isDeeplyObjectType(type) {
    return type.isIntersection()
        ? tsutils.intersectionTypeParts(type).every(tsutils.isObjectType)
        : tsutils.unionTypeParts(type).every(tsutils.isObjectType);
***REMOVED***
function isTypeFlagSetInUnion(type, flag) {
    return tsutils
        .unionTypeParts(type)
        .some(subType => tsutils.isTypeFlagSet(subType, flag));
***REMOVED***
