// @typescript-eslint/eslint-plugin/dist/rules/type-annotation-spacing.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; ***REMOVED*** ***REMOVED***;
    ***REMOVED***
    Object.defineProperty(o, k2, desc);
***REMOVED***) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
***REMOVED***));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v ***REMOVED***);
***REMOVED***) : function(o, v) {
    o["default"] = v;
***REMOVED***);
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {***REMOVED***;
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const util = __importStar(require("../util"));
const util_1 = require("../util");
const definition = {
    type: 'object',
    properties: {
        before: { type: 'boolean' ***REMOVED***,
        after: { type: 'boolean' ***REMOVED***,
    ***REMOVED***,
    additionalProperties: false,
***REMOVED***;
function createRules(options) {
    var _a;
    const globals = Object.assign(Object.assign({***REMOVED***, ((options === null || options === void 0 ? void 0 : options.before) !== undefined ? { before: options.before ***REMOVED*** : {***REMOVED***)), ((options === null || options === void 0 ? void 0 : options.after) !== undefined ? { after: options.after ***REMOVED*** : {***REMOVED***));
    const override = (_a = options === null || options === void 0 ? void 0 : options.overrides) !== null && _a !== void 0 ? _a : {***REMOVED***;
    const colon = Object.assign(Object.assign({ before: false, after: true ***REMOVED***, globals), override === null || override === void 0 ? void 0 : override.colon);
    const arrow = Object.assign(Object.assign({ before: true, after: true ***REMOVED***, globals), override === null || override === void 0 ? void 0 : override.arrow);
    return {
        colon: colon,
        arrow: arrow,
        variable: Object.assign(Object.assign({***REMOVED***, colon), override === null || override === void 0 ? void 0 : override.variable),
        property: Object.assign(Object.assign({***REMOVED***, colon), override === null || override === void 0 ? void 0 : override.property),
        parameter: Object.assign(Object.assign({***REMOVED***, colon), override === null || override === void 0 ? void 0 : override.parameter),
        returnType: Object.assign(Object.assign({***REMOVED***, colon), override === null || override === void 0 ? void 0 : override.returnType),
    ***REMOVED***;
***REMOVED***
function getIdentifierRules(rules, node) {
    const scope = node === null || node === void 0 ? void 0 : node.parent;
    if ((0, util_1.isVariableDeclarator)(scope)) {
        return rules.variable;
    ***REMOVED***
    else if ((0, util_1.isFunctionOrFunctionType)(scope)) {
        return rules.parameter;
    ***REMOVED***
    else {
        return rules.colon;
    ***REMOVED***
***REMOVED***
function getRules(rules, node) {
    var _a;
    const scope = (_a = node === null || node === void 0 ? void 0 : node.parent) === null || _a === void 0 ? void 0 : _a.parent;
    if ((0, util_1.isTSFunctionType)(scope) || (0, util_1.isTSConstructorType)(scope)) {
        return rules.arrow;
    ***REMOVED***
    else if ((0, util_1.isIdentifier)(scope)) {
        return getIdentifierRules(rules, scope);
    ***REMOVED***
    else if ((0, util_1.isClassOrTypeElement)(scope)) {
        return rules.property;
    ***REMOVED***
    else if ((0, util_1.isFunction)(scope)) {
        return rules.returnType;
    ***REMOVED***
    else {
        return rules.colon;
    ***REMOVED***
***REMOVED***
exports.default = util.createRule({
    name: 'type-annotation-spacing',
    meta: {
        type: 'layout',
        docs: {
            description: 'Require consistent spacing around type annotations',
            recommended: false,
        ***REMOVED***,
        fixable: 'whitespace',
        messages: {
            expectedSpaceAfter: "Expected a space after the '{{type***REMOVED******REMOVED***'.",
            expectedSpaceBefore: "Expected a space before the '{{type***REMOVED******REMOVED***'.",
            unexpectedSpaceAfter: "Unexpected space after the '{{type***REMOVED******REMOVED***'.",
            unexpectedSpaceBefore: "Unexpected space before the '{{type***REMOVED******REMOVED***'.",
            unexpectedSpaceBetween: "Unexpected space between the '{{previousToken***REMOVED******REMOVED***' and the '{{type***REMOVED******REMOVED***'.",
        ***REMOVED***,
        schema: [
            {
                type: 'object',
                properties: {
                    before: { type: 'boolean' ***REMOVED***,
                    after: { type: 'boolean' ***REMOVED***,
                    overrides: {
                        type: 'object',
                        properties: {
                            colon: definition,
                            arrow: definition,
                            variable: definition,
                            parameter: definition,
                            property: definition,
                            returnType: definition,
                        ***REMOVED***,
                        additionalProperties: false,
                    ***REMOVED***,
                ***REMOVED***,
                additionalProperties: false,
            ***REMOVED***,
        ],
    ***REMOVED***,
    defaultOptions: [
        // technically there is a default, but the overrides mean
        // that if we apply them here, it will break the no override case.
        {***REMOVED***,
    ],
    create(context, [options]) {
        const punctuators = [':', '=>'];
        const sourceCode = context.getSourceCode();
        const ruleSet = createRules(options);
        /**
         * Checks if there's proper spacing around type annotations (no space
         * before colon, one space after).
         */
        function checkTypeAnnotationSpacing(typeAnnotation) {
            const nextToken = typeAnnotation;
            const punctuatorTokenEnd = sourceCode.getTokenBefore(nextToken);
            let punctuatorTokenStart = punctuatorTokenEnd;
            let previousToken = sourceCode.getTokenBefore(punctuatorTokenEnd);
            let type = punctuatorTokenEnd.value;
            if (!punctuators.includes(type)) {
                return;
            ***REMOVED***
            const { before, after ***REMOVED*** = getRules(ruleSet, typeAnnotation);
            if (type === ':' && previousToken.value === '?') {
                if (
                // eslint-disable-next-line deprecation/deprecation -- TODO - switch once our min ESLint version is 6.7.0
                sourceCode.isSpaceBetweenTokens(previousToken, punctuatorTokenStart)) {
                    context.report({
                        node: punctuatorTokenStart,
                        messageId: 'unexpectedSpaceBetween',
                        data: {
                            type,
                            previousToken: previousToken.value,
                        ***REMOVED***,
                        fix(fixer) {
                            return fixer.removeRange([
                                previousToken.range[1],
                                punctuatorTokenStart.range[0],
                            ]);
                        ***REMOVED***,
                    ***REMOVED***);
                ***REMOVED***
                // shift the start to the ?
                type = '?:';
                punctuatorTokenStart = previousToken;
                previousToken = sourceCode.getTokenBefore(previousToken);
                // handle the +/- modifiers for optional modification operators
                if (previousToken.value === '+' || previousToken.value === '-') {
                    type = `${previousToken.value***REMOVED***?:`;
                    punctuatorTokenStart = previousToken;
                    previousToken = sourceCode.getTokenBefore(previousToken);
                ***REMOVED***
            ***REMOVED***
            const previousDelta = punctuatorTokenStart.range[0] - previousToken.range[1];
            const nextDelta = nextToken.range[0] - punctuatorTokenEnd.range[1];
            if (after && nextDelta === 0) {
                context.report({
                    node: punctuatorTokenEnd,
                    messageId: 'expectedSpaceAfter',
                    data: {
                        type,
                    ***REMOVED***,
                    fix(fixer) {
                        return fixer.insertTextAfter(punctuatorTokenEnd, ' ');
                    ***REMOVED***,
                ***REMOVED***);
            ***REMOVED***
            else if (!after && nextDelta > 0) {
                context.report({
                    node: punctuatorTokenEnd,
                    messageId: 'unexpectedSpaceAfter',
                    data: {
                        type,
                    ***REMOVED***,
                    fix(fixer) {
                        return fixer.removeRange([
                            punctuatorTokenEnd.range[1],
                            nextToken.range[0],
                        ]);
                    ***REMOVED***,
                ***REMOVED***);
            ***REMOVED***
            if (before && previousDelta === 0) {
                context.report({
                    node: punctuatorTokenStart,
                    messageId: 'expectedSpaceBefore',
                    data: {
                        type,
                    ***REMOVED***,
                    fix(fixer) {
                        return fixer.insertTextAfter(previousToken, ' ');
                    ***REMOVED***,
                ***REMOVED***);
            ***REMOVED***
            else if (!before && previousDelta > 0) {
                context.report({
                    node: punctuatorTokenStart,
                    messageId: 'unexpectedSpaceBefore',
                    data: {
                        type,
                    ***REMOVED***,
                    fix(fixer) {
                        return fixer.removeRange([
                            previousToken.range[1],
                            punctuatorTokenStart.range[0],
                        ]);
                    ***REMOVED***,
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***
        return {
            TSMappedType(node) {
                if (node.typeAnnotation) {
                    checkTypeAnnotationSpacing(node.typeAnnotation);
                ***REMOVED***
            ***REMOVED***,
            TSTypeAnnotation(node) {
                checkTypeAnnotationSpacing(node.typeAnnotation);
            ***REMOVED***,
        ***REMOVED***;
    ***REMOVED***,
***REMOVED***);
