// @typescript-eslint/eslint-plugin/dist/rules/typedef.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; ***REMOVED*** ***REMOVED***;
    ***REMOVED***
    Object.defineProperty(o, k2, desc);
***REMOVED***) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
***REMOVED***));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v ***REMOVED***);
***REMOVED***) : function(o, v) {
    o["default"] = v;
***REMOVED***);
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {***REMOVED***;
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const utils_1 = require("@typescript-eslint/utils");
const util = __importStar(require("../util"));
exports.default = util.createRule({
    name: 'typedef',
    meta: {
        docs: {
            description: 'Require type annotations in certain places',
            recommended: false,
        ***REMOVED***,
        messages: {
            expectedTypedef: 'Expected a type annotation.',
            expectedTypedefNamed: 'Expected {{name***REMOVED******REMOVED*** to have a type annotation.',
        ***REMOVED***,
        schema: [
            {
                type: 'object',
                properties: {
                    ["arrayDestructuring" /* OptionKeys.ArrayDestructuring */]: { type: 'boolean' ***REMOVED***,
                    ["arrowParameter" /* OptionKeys.ArrowParameter */]: { type: 'boolean' ***REMOVED***,
                    ["memberVariableDeclaration" /* OptionKeys.MemberVariableDeclaration */]: { type: 'boolean' ***REMOVED***,
                    ["objectDestructuring" /* OptionKeys.ObjectDestructuring */]: { type: 'boolean' ***REMOVED***,
                    ["parameter" /* OptionKeys.Parameter */]: { type: 'boolean' ***REMOVED***,
                    ["propertyDeclaration" /* OptionKeys.PropertyDeclaration */]: { type: 'boolean' ***REMOVED***,
                    ["variableDeclaration" /* OptionKeys.VariableDeclaration */]: { type: 'boolean' ***REMOVED***,
                    ["variableDeclarationIgnoreFunction" /* OptionKeys.VariableDeclarationIgnoreFunction */]: { type: 'boolean' ***REMOVED***,
                ***REMOVED***,
            ***REMOVED***,
        ],
        type: 'suggestion',
    ***REMOVED***,
    defaultOptions: [
        {
            ["arrayDestructuring" /* OptionKeys.ArrayDestructuring */]: false,
            ["arrowParameter" /* OptionKeys.ArrowParameter */]: false,
            ["memberVariableDeclaration" /* OptionKeys.MemberVariableDeclaration */]: false,
            ["objectDestructuring" /* OptionKeys.ObjectDestructuring */]: false,
            ["parameter" /* OptionKeys.Parameter */]: false,
            ["propertyDeclaration" /* OptionKeys.PropertyDeclaration */]: false,
            ["variableDeclaration" /* OptionKeys.VariableDeclaration */]: false,
            ["variableDeclarationIgnoreFunction" /* OptionKeys.VariableDeclarationIgnoreFunction */]: false,
        ***REMOVED***,
    ],
    create(context, [{ arrayDestructuring, arrowParameter, memberVariableDeclaration, objectDestructuring, parameter, propertyDeclaration, variableDeclaration, variableDeclarationIgnoreFunction, ***REMOVED***,]) {
        function report(location, name) {
            context.report({
                node: location,
                messageId: name ? 'expectedTypedefNamed' : 'expectedTypedef',
                data: { name ***REMOVED***,
            ***REMOVED***);
        ***REMOVED***
        function getNodeName(node) {
            return node.type === utils_1.AST_NODE_TYPES.Identifier ? node.name : undefined;
        ***REMOVED***
        function isForOfStatementContext(node) {
            let current = node.parent;
            while (current) {
                switch (current.type) {
                    case utils_1.AST_NODE_TYPES.VariableDeclarator:
                    case utils_1.AST_NODE_TYPES.VariableDeclaration:
                    case utils_1.AST_NODE_TYPES.ObjectPattern:
                    case utils_1.AST_NODE_TYPES.ArrayPattern:
                    case utils_1.AST_NODE_TYPES.Property:
                        current = current.parent;
                        break;
                    case utils_1.AST_NODE_TYPES.ForOfStatement:
                        return true;
                    default:
                        current = undefined;
                ***REMOVED***
            ***REMOVED***
            return false;
        ***REMOVED***
        function checkParameters(params) {
            for (const param of params) {
                let annotationNode;
                switch (param.type) {
                    case utils_1.AST_NODE_TYPES.AssignmentPattern:
                        annotationNode = param.left;
                        break;
                    case utils_1.AST_NODE_TYPES.TSParameterProperty:
                        annotationNode = param.parameter;
                        // Check TS parameter property with default value like `constructor(private param: string = 'something') {***REMOVED***`
                        if (annotationNode &&
                            annotationNode.type === utils_1.AST_NODE_TYPES.AssignmentPattern) {
                            annotationNode = annotationNode.left;
                        ***REMOVED***
                        break;
                    default:
                        annotationNode = param;
                        break;
                ***REMOVED***
                if (annotationNode !== undefined && !annotationNode.typeAnnotation) {
                    report(param, getNodeName(param));
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        function isVariableDeclarationIgnoreFunction(node) {
            return (variableDeclarationIgnoreFunction === true &&
                (node.type === utils_1.AST_NODE_TYPES.ArrowFunctionExpression ||
                    node.type === utils_1.AST_NODE_TYPES.FunctionExpression));
        ***REMOVED***
        function isAncestorHasTypeAnnotation(node) {
            let ancestor = node.parent;
            while (ancestor) {
                if ((ancestor.type === utils_1.AST_NODE_TYPES.ObjectPattern ||
                    ancestor.type === utils_1.AST_NODE_TYPES.ArrayPattern) &&
                    ancestor.typeAnnotation) {
                    return true;
                ***REMOVED***
                ancestor = ancestor.parent;
            ***REMOVED***
            return false;
        ***REMOVED***
        return Object.assign(Object.assign(Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({***REMOVED***, (arrayDestructuring && {
            ArrayPattern(node) {
                var _a, _b;
                if (((_a = node.parent) === null || _a === void 0 ? void 0 : _a.type) === utils_1.AST_NODE_TYPES.RestElement &&
                    node.parent.typeAnnotation) {
                    return;
                ***REMOVED***
                if (!node.typeAnnotation &&
                    !isForOfStatementContext(node) &&
                    !isAncestorHasTypeAnnotation(node) &&
                    ((_b = node.parent) === null || _b === void 0 ? void 0 : _b.type) !== utils_1.AST_NODE_TYPES.AssignmentExpression) {
                    report(node);
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***)), (arrowParameter && {
            ArrowFunctionExpression(node) {
                checkParameters(node.params);
            ***REMOVED***,
        ***REMOVED***)), (memberVariableDeclaration && {
            PropertyDefinition(node) {
                if (!(node.value && isVariableDeclarationIgnoreFunction(node.value)) &&
                    !node.typeAnnotation) {
                    report(node, node.key.type === utils_1.AST_NODE_TYPES.Identifier
                        ? node.key.name
                        : undefined);
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***)), (parameter && {
            'FunctionDeclaration, FunctionExpression'(node) {
                checkParameters(node.params);
            ***REMOVED***,
        ***REMOVED***)), (objectDestructuring && {
            ObjectPattern(node) {
                if (!node.typeAnnotation &&
                    !isForOfStatementContext(node) &&
                    !isAncestorHasTypeAnnotation(node)) {
                    report(node);
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***)), (propertyDeclaration && {
            'TSIndexSignature, TSPropertySignature'(node) {
                if (!node.typeAnnotation) {
                    report(node, node.type === utils_1.AST_NODE_TYPES.TSPropertySignature
                        ? getNodeName(node.key)
                        : undefined);
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***)), { VariableDeclarator(node) {
                if (!variableDeclaration ||
                    node.id.typeAnnotation ||
                    (node.id.type === utils_1.AST_NODE_TYPES.ArrayPattern &&
                        !arrayDestructuring) ||
                    (node.id.type === utils_1.AST_NODE_TYPES.ObjectPattern &&
                        !objectDestructuring) ||
                    (node.init && isVariableDeclarationIgnoreFunction(node.init))) {
                    return;
                ***REMOVED***
                let current = node.parent;
                while (current) {
                    switch (current.type) {
                        case utils_1.AST_NODE_TYPES.VariableDeclaration:
                            // Keep looking upwards
                            current = current.parent;
                            break;
                        case utils_1.AST_NODE_TYPES.ForOfStatement:
                        case utils_1.AST_NODE_TYPES.ForInStatement:
                            // Stop traversing and don't report an error
                            return;
                        default:
                            // Stop traversing
                            current = undefined;
                            break;
                    ***REMOVED***
                ***REMOVED***
                report(node, getNodeName(node.id));
            ***REMOVED*** ***REMOVED***);
    ***REMOVED***,
***REMOVED***);
