// css-select/lib/compile.js
/*
	compiles a selector to an executable function
*/

module.exports = compile;

var parse = require("css-what").parse;
var BaseFuncs = require("boolbase");
var sortRules = require("./sort.js");
var procedure = require("./procedure.json");
var Rules = require("./general.js");
var Pseudos = require("./pseudos.js");
var trueFunc = BaseFuncs.trueFunc;
var falseFunc = BaseFuncs.falseFunc;

var filters = Pseudos.filters;

function compile(selector, options, context) {
    var next = compileUnsafe(selector, options, context);
    return wrap(next, options);
***REMOVED***

function wrap(next, options) {
    var adapter = options.adapter;

    return function base(elem) {
        return adapter.isTag(elem) && next(elem);
    ***REMOVED***;
***REMOVED***

function compileUnsafe(selector, options, context) {
    var token = parse(selector, options);
    return compileToken(token, options, context);
***REMOVED***

function includesScopePseudo(t) {
    return (
        t.type === "pseudo" &&
        (t.name === "scope" ||
            (Array.isArray(t.data) &&
                t.data.some(function(data) {
                    return data.some(includesScopePseudo);
                ***REMOVED***)))
    );
***REMOVED***

var DESCENDANT_TOKEN = { type: "descendant" ***REMOVED***;
var FLEXIBLE_DESCENDANT_TOKEN = { type: "_flexibleDescendant" ***REMOVED***;
var SCOPE_TOKEN = { type: "pseudo", name: "scope" ***REMOVED***;
var PLACEHOLDER_ELEMENT = {***REMOVED***;

//CSS 4 Spec (Draft): 3.3.1. Absolutizing a Scope-relative Selector
//http://www.w3.org/TR/selectors4/#absolutizing
function absolutize(token, options, context) {
    var adapter = options.adapter;

    //TODO better check if context is document
    var hasContext =
        !!context &&
        !!context.length &&
        context.every(function(e) {
            return e === PLACEHOLDER_ELEMENT || !!adapter.getParent(e);
        ***REMOVED***);

    token.forEach(function(t) {
        if (t.length > 0 && isTraversal(t[0]) && t[0].type !== "descendant") {
            //don't return in else branch
        ***REMOVED*** else if (hasContext && !(Array.isArray(t) ? t.some(includesScopePseudo) : includesScopePseudo(t))) {
            t.unshift(DESCENDANT_TOKEN);
        ***REMOVED*** else {
            return;
        ***REMOVED***

        t.unshift(SCOPE_TOKEN);
    ***REMOVED***);
***REMOVED***

function compileToken(token, options, context) {
    token = token.filter(function(t) {
        return t.length > 0;
    ***REMOVED***);

    token.forEach(sortRules);

    var isArrayContext = Array.isArray(context);

    context = (options && options.context) || context;

    if (context && !isArrayContext) context = [context];

    absolutize(token, options, context);

    var shouldTestNextSiblings = false;

    var query = token
        .map(function(rules) {
            if (rules[0] && rules[1] && rules[0].name === "scope") {
                var ruleType = rules[1].type;
                if (isArrayContext && ruleType === "descendant") {
                    rules[1] = FLEXIBLE_DESCENDANT_TOKEN;
                ***REMOVED*** else if (ruleType === "adjacent" || ruleType === "sibling") {
                    shouldTestNextSiblings = true;
                ***REMOVED***
            ***REMOVED***
            return compileRules(rules, options, context);
        ***REMOVED***)
        .reduce(reduceRules, falseFunc);

    query.shouldTestNextSiblings = shouldTestNextSiblings;

    return query;
***REMOVED***

function isTraversal(t) {
    return procedure[t.type] < 0;
***REMOVED***

function compileRules(rules, options, context) {
    return rules.reduce(function(func, rule) {
        if (func === falseFunc) return func;

        if (!(rule.type in Rules)) {
            throw new Error("Rule type " + rule.type + " is not supported by css-select");
        ***REMOVED***

        return Rules[rule.type](func, rule, options, context);
    ***REMOVED***, (options && options.rootFunc) || trueFunc);
***REMOVED***

function reduceRules(a, b) {
    if (b === falseFunc || a === trueFunc) {
        return a;
    ***REMOVED***
    if (a === falseFunc || b === trueFunc) {
        return b;
    ***REMOVED***

    return function combine(elem) {
        return a(elem) || b(elem);
    ***REMOVED***;
***REMOVED***

function containsTraversal(t) {
    return t.some(isTraversal);
***REMOVED***

//:not, :has and :matches have to compile selectors
//doing this in lib/pseudos.js would lead to circular dependencies,
//so we add them here
filters.not = function(next, token, options, context) {
    var opts = {
        xmlMode: !!(options && options.xmlMode),
        strict: !!(options && options.strict),
        adapter: options.adapter
    ***REMOVED***;

    if (opts.strict) {
        if (token.length > 1 || token.some(containsTraversal)) {
            throw new Error("complex selectors in :not aren't allowed in strict mode");
        ***REMOVED***
    ***REMOVED***

    var func = compileToken(token, opts, context);

    if (func === falseFunc) return next;
    if (func === trueFunc) return falseFunc;

    return function not(elem) {
        return !func(elem) && next(elem);
    ***REMOVED***;
***REMOVED***;

filters.has = function(next, token, options) {
    var adapter = options.adapter;
    var opts = {
        xmlMode: !!(options && options.xmlMode),
        strict: !!(options && options.strict),
        adapter: adapter
    ***REMOVED***;

    //FIXME: Uses an array as a pointer to the current element (side effects)
    var context = token.some(containsTraversal) ? [PLACEHOLDER_ELEMENT] : null;

    var func = compileToken(token, opts, context);

    if (func === falseFunc) return falseFunc;
    if (func === trueFunc) {
        return function hasChild(elem) {
            return adapter.getChildren(elem).some(adapter.isTag) && next(elem);
        ***REMOVED***;
    ***REMOVED***

    func = wrap(func, options);

    if (context) {
        return function has(elem) {
            return next(elem) && ((context[0] = elem), adapter.existsOne(func, adapter.getChildren(elem)));
        ***REMOVED***;
    ***REMOVED***

    return function has(elem) {
        return next(elem) && adapter.existsOne(func, adapter.getChildren(elem));
    ***REMOVED***;
***REMOVED***;

filters.matches = function(next, token, options, context) {
    var opts = {
        xmlMode: !!(options && options.xmlMode),
        strict: !!(options && options.strict),
        rootFunc: next,
        adapter: options.adapter
    ***REMOVED***;

    return compileToken(token, opts, context);
***REMOVED***;

compile.compileToken = compileToken;
compile.compileUnsafe = compileUnsafe;
compile.Pseudos = Pseudos;
