// webpack/lib/optimize/ModuleConcatenationPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const asyncLib = require("neo-async");
const ChunkGraph = require("../ChunkGraph");
const ModuleGraph = require("../ModuleGraph");
const { JS_TYPE ***REMOVED*** = require("../ModuleSourceTypesConstants");
const { STAGE_DEFAULT ***REMOVED*** = require("../OptimizationStages");
const HarmonyImportDependency = require("../dependencies/HarmonyImportDependency");
const { compareModulesByIdentifier ***REMOVED*** = require("../util/comparators");
const {
	intersectRuntime,
	mergeRuntimeOwned,
	filterRuntime,
	runtimeToString,
	mergeRuntime
***REMOVED*** = require("../util/runtime");
const ConcatenatedModule = require("./ConcatenatedModule");

/** @typedef {import("../Compilation")***REMOVED*** Compilation */
/** @typedef {import("../Compiler")***REMOVED*** Compiler */
/** @typedef {import("../Module")***REMOVED*** Module */
/** @typedef {import("../Module").BuildInfo***REMOVED*** BuildInfo */
/** @typedef {import("../RequestShortener")***REMOVED*** RequestShortener */
/** @typedef {import("../util/runtime").RuntimeSpec***REMOVED*** RuntimeSpec */

/**
 * @typedef {object***REMOVED*** Statistics
 * @property {number***REMOVED*** cached
 * @property {number***REMOVED*** alreadyInConfig
 * @property {number***REMOVED*** invalidModule
 * @property {number***REMOVED*** incorrectChunks
 * @property {number***REMOVED*** incorrectDependency
 * @property {number***REMOVED*** incorrectModuleDependency
 * @property {number***REMOVED*** incorrectChunksOfImporter
 * @property {number***REMOVED*** incorrectRuntimeCondition
 * @property {number***REMOVED*** importerFailed
 * @property {number***REMOVED*** added
 */

/**
 * @param {string***REMOVED*** msg message
 * @returns {string***REMOVED*** formatted message
 */
const formatBailoutReason = msg => `ModuleConcatenation bailout: ${msg***REMOVED***`;

const PLUGIN_NAME = "ModuleConcatenationPlugin";

class ModuleConcatenationPlugin {
	/**
	 * Apply the plugin
	 * @param {Compiler***REMOVED*** compiler the compiler instance
	 * @returns {void***REMOVED***
	 */
	apply(compiler) {
		const { _backCompat: backCompat ***REMOVED*** = compiler;
		compiler.hooks.compilation.tap(PLUGIN_NAME, compilation => {
			if (compilation.moduleMemCaches) {
				throw new Error(
					"optimization.concatenateModules can't be used with cacheUnaffected as module concatenation is a global effect"
				);
			***REMOVED***
			const moduleGraph = compilation.moduleGraph;
			/** @type {Map<Module, string | ((requestShortener: RequestShortener) => string)>***REMOVED*** */
			const bailoutReasonMap = new Map();

			/**
			 * @param {Module***REMOVED*** module the module
			 * @param {string | ((requestShortener: RequestShortener) => string)***REMOVED*** reason the reason
			 */
			const setBailoutReason = (module, reason) => {
				setInnerBailoutReason(module, reason);
				moduleGraph
					.getOptimizationBailout(module)
					.push(
						typeof reason === "function"
							? rs => formatBailoutReason(reason(rs))
							: formatBailoutReason(reason)
					);
			***REMOVED***;

			/**
			 * @param {Module***REMOVED*** module the module
			 * @param {string | ((requestShortener: RequestShortener) => string)***REMOVED*** reason the reason
			 */
			const setInnerBailoutReason = (module, reason) => {
				bailoutReasonMap.set(module, reason);
			***REMOVED***;

			/**
			 * @param {Module***REMOVED*** module the module
			 * @param {RequestShortener***REMOVED*** requestShortener the request shortener
			 * @returns {string | ((requestShortener: RequestShortener) => string) | undefined***REMOVED*** the reason
			 */
			const getInnerBailoutReason = (module, requestShortener) => {
				const reason = bailoutReasonMap.get(module);
				if (typeof reason === "function") return reason(requestShortener);
				return reason;
			***REMOVED***;

			/**
			 * @param {Module***REMOVED*** module the module
			 * @param {Module | ((requestShortener: RequestShortener) => string)***REMOVED*** problem the problem
			 * @returns {(requestShortener: RequestShortener) => string***REMOVED*** the reason
			 */
			const formatBailoutWarning = (module, problem) => requestShortener => {
				if (typeof problem === "function") {
					return formatBailoutReason(
						`Cannot concat with ${module.readableIdentifier(
							requestShortener
						)***REMOVED***: ${problem(requestShortener)***REMOVED***`
					);
				***REMOVED***
				const reason = getInnerBailoutReason(module, requestShortener);
				const reasonWithPrefix = reason ? `: ${reason***REMOVED***` : "";
				if (module === problem) {
					return formatBailoutReason(
						`Cannot concat with ${module.readableIdentifier(
							requestShortener
						)***REMOVED***${reasonWithPrefix***REMOVED***`
					);
				***REMOVED***
				return formatBailoutReason(
					`Cannot concat with ${module.readableIdentifier(
						requestShortener
					)***REMOVED*** because of ${problem.readableIdentifier(
						requestShortener
					)***REMOVED***${reasonWithPrefix***REMOVED***`
				);
			***REMOVED***;

			compilation.hooks.optimizeChunkModules.tapAsync(
				{
					name: PLUGIN_NAME,
					stage: STAGE_DEFAULT
				***REMOVED***,
				(allChunks, modules, callback) => {
					const logger = compilation.getLogger(
						"webpack.ModuleConcatenationPlugin"
					);
					const { chunkGraph, moduleGraph ***REMOVED*** = compilation;
					const relevantModules = [];
					const possibleInners = new Set();
					const context = {
						chunkGraph,
						moduleGraph
					***REMOVED***;
					logger.time("select relevant modules");
					for (const module of modules) {
						let canBeRoot = true;
						let canBeInner = true;

						const bailoutReason = module.getConcatenationBailoutReason(context);
						if (bailoutReason) {
							setBailoutReason(module, bailoutReason);
							continue;
						***REMOVED***

						// Must not be an async module
						if (moduleGraph.isAsync(module)) {
							setBailoutReason(module, "Module is async");
							continue;
						***REMOVED***

						// Must be in strict mode
						if (!(/** @type {BuildInfo***REMOVED*** */ (module.buildInfo).strict)) {
							setBailoutReason(module, "Module is not in strict mode");
							continue;
						***REMOVED***

						// Module must be in any chunk (we don't want to do useless work)
						if (chunkGraph.getNumberOfModuleChunks(module) === 0) {
							setBailoutReason(module, "Module is not in any chunk");
							continue;
						***REMOVED***

						// Exports must be known (and not dynamic)
						const exportsInfo = moduleGraph.getExportsInfo(module);
						const relevantExports = exportsInfo.getRelevantExports(undefined);
						const unknownReexports = relevantExports.filter(
							exportInfo =>
								exportInfo.isReexport() && !exportInfo.getTarget(moduleGraph)
						);
						if (unknownReexports.length > 0) {
							setBailoutReason(
								module,
								`Reexports in this module do not have a static target (${Array.from(
									unknownReexports,
									exportInfo =>
										`${
											exportInfo.name || "other exports"
										***REMOVED***: ${exportInfo.getUsedInfo()***REMOVED***`
								).join(", ")***REMOVED***)`
							);
							continue;
						***REMOVED***

						// Root modules must have a static list of exports
						const unknownProvidedExports = relevantExports.filter(
							exportInfo => exportInfo.provided !== true
						);
						if (unknownProvidedExports.length > 0) {
							setBailoutReason(
								module,
								`List of module exports is dynamic (${Array.from(
									unknownProvidedExports,
									exportInfo =>
										`${
											exportInfo.name || "other exports"
										***REMOVED***: ${exportInfo.getProvidedInfo()***REMOVED*** and ${exportInfo.getUsedInfo()***REMOVED***`
								).join(", ")***REMOVED***)`
							);
							canBeRoot = false;
						***REMOVED***

						// Module must not be an entry point
						if (chunkGraph.isEntryModule(module)) {
							setInnerBailoutReason(module, "Module is an entry point");
							canBeInner = false;
						***REMOVED***

						if (canBeRoot) relevantModules.push(module);
						if (canBeInner) possibleInners.add(module);
					***REMOVED***
					logger.timeEnd("select relevant modules");
					logger.debug(
						`${relevantModules.length***REMOVED*** potential root modules, ${possibleInners.size***REMOVED*** potential inner modules`
					);
					// sort by depth
					// modules with lower depth are more likely suited as roots
					// this improves performance, because modules already selected as inner are skipped
					logger.time("sort relevant modules");
					relevantModules.sort(
						(a, b) =>
							/** @type {number***REMOVED*** */ (moduleGraph.getDepth(a)) -
							/** @type {number***REMOVED*** */ (moduleGraph.getDepth(b))
					);
					logger.timeEnd("sort relevant modules");

					/** @type {Statistics***REMOVED*** */
					const stats = {
						cached: 0,
						alreadyInConfig: 0,
						invalidModule: 0,
						incorrectChunks: 0,
						incorrectDependency: 0,
						incorrectModuleDependency: 0,
						incorrectChunksOfImporter: 0,
						incorrectRuntimeCondition: 0,
						importerFailed: 0,
						added: 0
					***REMOVED***;
					let statsCandidates = 0;
					let statsSizeSum = 0;
					let statsEmptyConfigurations = 0;

					logger.time("find modules to concatenate");
					const concatConfigurations = [];
					const usedAsInner = new Set();
					for (const currentRoot of relevantModules) {
						// when used by another configuration as inner:
						// the other configuration is better and we can skip this one
						// TODO reconsider that when it's only used in a different runtime
						if (usedAsInner.has(currentRoot)) continue;

						let chunkRuntime;
						for (const r of chunkGraph.getModuleRuntimes(currentRoot)) {
							chunkRuntime = mergeRuntimeOwned(chunkRuntime, r);
						***REMOVED***
						const exportsInfo = moduleGraph.getExportsInfo(currentRoot);
						const filteredRuntime = filterRuntime(chunkRuntime, r =>
							exportsInfo.isModuleUsed(r)
						);
						const activeRuntime =
							filteredRuntime === true
								? chunkRuntime
								: filteredRuntime === false
									? undefined
									: filteredRuntime;

						// create a configuration with the root
						const currentConfiguration = new ConcatConfiguration(
							currentRoot,
							activeRuntime
						);

						// cache failures to add modules
						const failureCache = new Map();

						// potential optional import candidates
						/** @type {Set<Module>***REMOVED*** */
						const candidates = new Set();

						// try to add all imports
						for (const imp of this._getImports(
							compilation,
							currentRoot,
							activeRuntime
						)) {
							candidates.add(imp);
						***REMOVED***

						for (const imp of candidates) {
							const impCandidates = new Set();
							const problem = this._tryToAdd(
								compilation,
								currentConfiguration,
								imp,
								chunkRuntime,
								activeRuntime,
								possibleInners,
								impCandidates,
								failureCache,
								chunkGraph,
								true,
								stats
							);
							if (problem) {
								failureCache.set(imp, problem);
								currentConfiguration.addWarning(imp, problem);
							***REMOVED*** else {
								for (const c of impCandidates) {
									candidates.add(c);
								***REMOVED***
							***REMOVED***
						***REMOVED***
						statsCandidates += candidates.size;
						if (!currentConfiguration.isEmpty()) {
							const modules = currentConfiguration.getModules();
							statsSizeSum += modules.size;
							concatConfigurations.push(currentConfiguration);
							for (const module of modules) {
								if (module !== currentConfiguration.rootModule) {
									usedAsInner.add(module);
								***REMOVED***
							***REMOVED***
						***REMOVED*** else {
							statsEmptyConfigurations++;
							const optimizationBailouts =
								moduleGraph.getOptimizationBailout(currentRoot);
							for (const warning of currentConfiguration.getWarningsSorted()) {
								optimizationBailouts.push(
									formatBailoutWarning(warning[0], warning[1])
								);
							***REMOVED***
						***REMOVED***
					***REMOVED***
					logger.timeEnd("find modules to concatenate");
					logger.debug(
						`${
							concatConfigurations.length
						***REMOVED*** successful concat configurations (avg size: ${
							statsSizeSum / concatConfigurations.length
						***REMOVED***), ${statsEmptyConfigurations***REMOVED*** bailed out completely`
					);
					logger.debug(
						`${statsCandidates***REMOVED*** candidates were considered for adding (${stats.cached***REMOVED*** cached failure, ${stats.alreadyInConfig***REMOVED*** already in config, ${stats.invalidModule***REMOVED*** invalid module, ${stats.incorrectChunks***REMOVED*** incorrect chunks, ${stats.incorrectDependency***REMOVED*** incorrect dependency, ${stats.incorrectChunksOfImporter***REMOVED*** incorrect chunks of importer, ${stats.incorrectModuleDependency***REMOVED*** incorrect module dependency, ${stats.incorrectRuntimeCondition***REMOVED*** incorrect runtime condition, ${stats.importerFailed***REMOVED*** importer failed, ${stats.added***REMOVED*** added)`
					);
					// HACK: Sort configurations by length and start with the longest one
					// to get the biggest groups possible. Used modules are marked with usedModules
					// TODO: Allow to reuse existing configuration while trying to add dependencies.
					// This would improve performance. O(n^2) -> O(n)
					logger.time("sort concat configurations");
					concatConfigurations.sort((a, b) => b.modules.size - a.modules.size);
					logger.timeEnd("sort concat configurations");
					const usedModules = new Set();

					logger.time("create concatenated modules");
					asyncLib.each(
						concatConfigurations,
						(concatConfiguration, callback) => {
							const rootModule = concatConfiguration.rootModule;

							// Avoid overlapping configurations
							// TODO: remove this when todo above is fixed
							if (usedModules.has(rootModule)) return callback();
							const modules = concatConfiguration.getModules();
							for (const m of modules) {
								usedModules.add(m);
							***REMOVED***

							// Create a new ConcatenatedModule
							ConcatenatedModule.getCompilationHooks(compilation);
							const newModule = ConcatenatedModule.create(
								rootModule,
								modules,
								concatConfiguration.runtime,
								compilation,
								compiler.root,
								compilation.outputOptions.hashFunction
							);

							const build = () => {
								newModule.build(
									compiler.options,
									compilation,
									/** @type {EXPECTED_ANY***REMOVED*** */
									(null),
									/** @type {EXPECTED_ANY***REMOVED*** */
									(null),
									err => {
										if (err) {
											if (!err.module) {
												err.module = newModule;
											***REMOVED***
											return callback(err);
										***REMOVED***
										integrate();
									***REMOVED***
								);
							***REMOVED***;

							const integrate = () => {
								if (backCompat) {
									ChunkGraph.setChunkGraphForModule(newModule, chunkGraph);
									ModuleGraph.setModuleGraphForModule(newModule, moduleGraph);
								***REMOVED***

								for (const warning of concatConfiguration.getWarningsSorted()) {
									moduleGraph
										.getOptimizationBailout(newModule)
										.push(formatBailoutWarning(warning[0], warning[1]));
								***REMOVED***
								moduleGraph.cloneModuleAttributes(rootModule, newModule);
								for (const m of modules) {
									// add to builtModules when one of the included modules was built
									if (compilation.builtModules.has(m)) {
										compilation.builtModules.add(newModule);
									***REMOVED***
									if (m !== rootModule) {
										// attach external references to the concatenated module too
										moduleGraph.copyOutgoingModuleConnections(
											m,
											newModule,
											c =>
												c.originModule === m &&
												!(
													c.dependency instanceof HarmonyImportDependency &&
													modules.has(c.module)
												)
										);
										// remove module from chunk
										for (const chunk of chunkGraph.getModuleChunksIterable(
											rootModule
										)) {
											const sourceTypes = chunkGraph.getChunkModuleSourceTypes(
												chunk,
												m
											);
											if (sourceTypes.size === 1) {
												chunkGraph.disconnectChunkAndModule(chunk, m);
											***REMOVED*** else {
												const newSourceTypes = new Set(sourceTypes);
												newSourceTypes.delete(JS_TYPE);
												chunkGraph.setChunkModuleSourceTypes(
													chunk,
													m,
													newSourceTypes
												);
											***REMOVED***
										***REMOVED***
									***REMOVED***
								***REMOVED***
								compilation.modules.delete(rootModule);
								ChunkGraph.clearChunkGraphForModule(rootModule);
								ModuleGraph.clearModuleGraphForModule(rootModule);

								// remove module from chunk
								chunkGraph.replaceModule(rootModule, newModule);
								// replace module references with the concatenated module
								moduleGraph.moveModuleConnections(rootModule, newModule, c => {
									const otherModule =
										c.module === rootModule ? c.originModule : c.module;
									const innerConnection =
										c.dependency instanceof HarmonyImportDependency &&
										modules.has(/** @type {Module***REMOVED*** */ (otherModule));
									return !innerConnection;
								***REMOVED***);
								// add concatenated module to the compilation
								compilation.modules.add(newModule);

								callback();
							***REMOVED***;

							build();
						***REMOVED***,
						err => {
							logger.timeEnd("create concatenated modules");
							process.nextTick(callback.bind(null, err));
						***REMOVED***
					);
				***REMOVED***
			);
		***REMOVED***);
	***REMOVED***

	/**
	 * @param {Compilation***REMOVED*** compilation the compilation
	 * @param {Module***REMOVED*** module the module to be added
	 * @param {RuntimeSpec***REMOVED*** runtime the runtime scope
	 * @returns {Set<Module>***REMOVED*** the imported modules
	 */
	_getImports(compilation, module, runtime) {
		const moduleGraph = compilation.moduleGraph;
		const set = new Set();
		for (const dep of module.dependencies) {
			// Get reference info only for harmony Dependencies
			if (!(dep instanceof HarmonyImportDependency)) continue;

			const connection = moduleGraph.getConnection(dep);
			// Reference is valid and has a module
			if (
				!connection ||
				!connection.module ||
				!connection.isTargetActive(runtime)
			) {
				continue;
			***REMOVED***

			const importedNames = compilation.getDependencyReferencedExports(
				dep,
				undefined
			);

			if (
				importedNames.every(i =>
					Array.isArray(i) ? i.length > 0 : i.name.length > 0
				) ||
				Array.isArray(moduleGraph.getProvidedExports(module))
			) {
				set.add(connection.module);
			***REMOVED***
		***REMOVED***
		return set;
	***REMOVED***

	/**
	 * @param {Compilation***REMOVED*** compilation webpack compilation
	 * @param {ConcatConfiguration***REMOVED*** config concat configuration (will be modified when added)
	 * @param {Module***REMOVED*** module the module to be added
	 * @param {RuntimeSpec***REMOVED*** runtime the runtime scope of the generated code
	 * @param {RuntimeSpec***REMOVED*** activeRuntime the runtime scope of the root module
	 * @param {Set<Module>***REMOVED*** possibleModules modules that are candidates
	 * @param {Set<Module>***REMOVED*** candidates list of potential candidates (will be added to)
	 * @param {Map<Module, Module | ((requestShortener: RequestShortener) => string)>***REMOVED*** failureCache cache for problematic modules to be more performant
	 * @param {ChunkGraph***REMOVED*** chunkGraph the chunk graph
	 * @param {boolean***REMOVED*** avoidMutateOnFailure avoid mutating the config when adding fails
	 * @param {Statistics***REMOVED*** statistics gathering metrics
	 * @returns {null | Module | ((requestShortener: RequestShortener) => string)***REMOVED*** the problematic module
	 */
	_tryToAdd(
		compilation,
		config,
		module,
		runtime,
		activeRuntime,
		possibleModules,
		candidates,
		failureCache,
		chunkGraph,
		avoidMutateOnFailure,
		statistics
	) {
		const cacheEntry = failureCache.get(module);
		if (cacheEntry) {
			statistics.cached++;
			return cacheEntry;
		***REMOVED***

		// Already added?
		if (config.has(module)) {
			statistics.alreadyInConfig++;
			return null;
		***REMOVED***

		// Not possible to add?
		if (!possibleModules.has(module)) {
			statistics.invalidModule++;
			failureCache.set(module, module); // cache failures for performance
			return module;
		***REMOVED***

		// Module must be in the correct chunks
		const missingChunks = Array.from(
			chunkGraph.getModuleChunksIterable(config.rootModule)
		).filter(chunk => !chunkGraph.isModuleInChunk(module, chunk));
		if (missingChunks.length > 0) {
			/**
			 * @param {RequestShortener***REMOVED*** requestShortener request shortener
			 * @returns {string***REMOVED*** problem description
			 */
			const problem = requestShortener => {
				const missingChunksList = Array.from(
					new Set(missingChunks.map(chunk => chunk.name || "unnamed chunk(s)"))
				).sort();
				const chunks = Array.from(
					new Set(
						Array.from(chunkGraph.getModuleChunksIterable(module)).map(
							chunk => chunk.name || "unnamed chunk(s)"
						)
					)
				).sort();
				return `Module ${module.readableIdentifier(
					requestShortener
				)***REMOVED*** is not in the same chunk(s) (expected in chunk(s) ${missingChunksList.join(
					", "
				)***REMOVED***, module is in chunk(s) ${chunks.join(", ")***REMOVED***)`;
			***REMOVED***;
			statistics.incorrectChunks++;
			failureCache.set(module, problem); // cache failures for performance
			return problem;
		***REMOVED***

		const moduleGraph = compilation.moduleGraph;

		const incomingConnections =
			moduleGraph.getIncomingConnectionsByOriginModule(module);

		const incomingConnectionsFromNonModules =
			incomingConnections.get(null) || incomingConnections.get(undefined);
		if (incomingConnectionsFromNonModules) {
			const activeNonModulesConnections =
				incomingConnectionsFromNonModules.filter(connection =>
					// We are not interested in inactive connections
					// or connections without dependency
					connection.isActive(runtime)
				);
			if (activeNonModulesConnections.length > 0) {
				/**
				 * @param {RequestShortener***REMOVED*** requestShortener request shortener
				 * @returns {string***REMOVED*** problem description
				 */
				const problem = requestShortener => {
					const importingExplanations = new Set(
						activeNonModulesConnections.map(c => c.explanation).filter(Boolean)
					);
					const explanations = Array.from(importingExplanations).sort();
					return `Module ${module.readableIdentifier(
						requestShortener
					)***REMOVED*** is referenced ${
						explanations.length > 0
							? `by: ${explanations.join(", ")***REMOVED***`
							: "in an unsupported way"
					***REMOVED***`;
				***REMOVED***;
				statistics.incorrectDependency++;
				failureCache.set(module, problem); // cache failures for performance
				return problem;
			***REMOVED***
		***REMOVED***

		/** @type {Map<Module, readonly ModuleGraph.ModuleGraphConnection[]>***REMOVED*** */
		const incomingConnectionsFromModules = new Map();
		for (const [originModule, connections] of incomingConnections) {
			if (originModule) {
				// Ignore connection from orphan modules
				if (chunkGraph.getNumberOfModuleChunks(originModule) === 0) continue;

				// We don't care for connections from other runtimes
				let originRuntime;
				for (const r of chunkGraph.getModuleRuntimes(originModule)) {
					originRuntime = mergeRuntimeOwned(originRuntime, r);
				***REMOVED***

				if (!intersectRuntime(runtime, originRuntime)) continue;

				// We are not interested in inactive connections
				const activeConnections = connections.filter(connection =>
					connection.isActive(runtime)
				);
				if (activeConnections.length > 0)
					incomingConnectionsFromModules.set(originModule, activeConnections);
			***REMOVED***
		***REMOVED***

		const incomingModules = Array.from(incomingConnectionsFromModules.keys());

		// Module must be in the same chunks like the referencing module
		const otherChunkModules = incomingModules.filter(originModule => {
			for (const chunk of chunkGraph.getModuleChunksIterable(
				config.rootModule
			)) {
				if (!chunkGraph.isModuleInChunk(originModule, chunk)) {
					return true;
				***REMOVED***
			***REMOVED***
			return false;
		***REMOVED***);
		if (otherChunkModules.length > 0) {
			/**
			 * @param {RequestShortener***REMOVED*** requestShortener request shortener
			 * @returns {string***REMOVED*** problem description
			 */
			const problem = requestShortener => {
				const names = otherChunkModules
					.map(m => m.readableIdentifier(requestShortener))
					.sort();
				return `Module ${module.readableIdentifier(
					requestShortener
				)***REMOVED*** is referenced from different chunks by these modules: ${names.join(
					", "
				)***REMOVED***`;
			***REMOVED***;
			statistics.incorrectChunksOfImporter++;
			failureCache.set(module, problem); // cache failures for performance
			return problem;
		***REMOVED***

		/** @type {Map<Module, readonly ModuleGraph.ModuleGraphConnection[]>***REMOVED*** */
		const nonHarmonyConnections = new Map();
		for (const [originModule, connections] of incomingConnectionsFromModules) {
			const selected = connections.filter(
				connection =>
					!connection.dependency ||
					!(connection.dependency instanceof HarmonyImportDependency)
			);
			if (selected.length > 0)
				nonHarmonyConnections.set(originModule, connections);
		***REMOVED***
		if (nonHarmonyConnections.size > 0) {
			/**
			 * @param {RequestShortener***REMOVED*** requestShortener request shortener
			 * @returns {string***REMOVED*** problem description
			 */
			const problem = requestShortener => {
				const names = Array.from(nonHarmonyConnections)
					.map(
						([originModule, connections]) =>
							`${originModule.readableIdentifier(
								requestShortener
							)***REMOVED*** (referenced with ${Array.from(
								new Set(
									connections
										.map(c => c.dependency && c.dependency.type)
										.filter(Boolean)
								)
							)
								.sort()
								.join(", ")***REMOVED***)`
					)
					.sort();
				return `Module ${module.readableIdentifier(
					requestShortener
				)***REMOVED*** is referenced from these modules with unsupported syntax: ${names.join(
					", "
				)***REMOVED***`;
			***REMOVED***;
			statistics.incorrectModuleDependency++;
			failureCache.set(module, problem); // cache failures for performance
			return problem;
		***REMOVED***

		if (runtime !== undefined && typeof runtime !== "string") {
			// Module must be consistently referenced in the same runtimes
			/** @type {{ originModule: Module, runtimeCondition: RuntimeSpec ***REMOVED***[]***REMOVED*** */
			const otherRuntimeConnections = [];
			outer: for (const [
				originModule,
				connections
			] of incomingConnectionsFromModules) {
				/** @type {false | RuntimeSpec***REMOVED*** */
				let currentRuntimeCondition = false;
				for (const connection of connections) {
					const runtimeCondition = filterRuntime(runtime, runtime =>
						connection.isTargetActive(runtime)
					);
					if (runtimeCondition === false) continue;
					if (runtimeCondition === true) continue outer;
					currentRuntimeCondition =
						currentRuntimeCondition !== false
							? mergeRuntime(currentRuntimeCondition, runtimeCondition)
							: runtimeCondition;
				***REMOVED***
				if (currentRuntimeCondition !== false) {
					otherRuntimeConnections.push({
						originModule,
						runtimeCondition: currentRuntimeCondition
					***REMOVED***);
				***REMOVED***
			***REMOVED***
			if (otherRuntimeConnections.length > 0) {
				/**
				 * @param {RequestShortener***REMOVED*** requestShortener request shortener
				 * @returns {string***REMOVED*** problem description
				 */
				const problem = requestShortener =>
					`Module ${module.readableIdentifier(
						requestShortener
					)***REMOVED*** is runtime-dependent referenced by these modules: ${Array.from(
						otherRuntimeConnections,
						({ originModule, runtimeCondition ***REMOVED***) =>
							`${originModule.readableIdentifier(
								requestShortener
							)***REMOVED*** (expected runtime ${runtimeToString(
								runtime
							)***REMOVED***, module is only referenced in ${runtimeToString(
								/** @type {RuntimeSpec***REMOVED*** */ (runtimeCondition)
							)***REMOVED***)`
					).join(", ")***REMOVED***`;
				statistics.incorrectRuntimeCondition++;
				failureCache.set(module, problem); // cache failures for performance
				return problem;
			***REMOVED***
		***REMOVED***

		let backup;
		if (avoidMutateOnFailure) {
			backup = config.snapshot();
		***REMOVED***

		// Add the module
		config.add(module);

		incomingModules.sort(compareModulesByIdentifier);

		// Every module which depends on the added module must be in the configuration too.
		for (const originModule of incomingModules) {
			const problem = this._tryToAdd(
				compilation,
				config,
				originModule,
				runtime,
				activeRuntime,
				possibleModules,
				candidates,
				failureCache,
				chunkGraph,
				false,
				statistics
			);
			if (problem) {
				if (backup !== undefined) config.rollback(backup);
				statistics.importerFailed++;
				failureCache.set(module, problem); // cache failures for performance
				return problem;
			***REMOVED***
		***REMOVED***

		// Add imports to possible candidates list
		for (const imp of this._getImports(compilation, module, runtime)) {
			candidates.add(imp);
		***REMOVED***
		statistics.added++;
		return null;
	***REMOVED***
***REMOVED***

/** @typedef {Module | ((requestShortener: RequestShortener) => string)***REMOVED*** Problem */

class ConcatConfiguration {
	/**
	 * @param {Module***REMOVED*** rootModule the root module
	 * @param {RuntimeSpec***REMOVED*** runtime the runtime
	 */
	constructor(rootModule, runtime) {
		this.rootModule = rootModule;
		this.runtime = runtime;
		/** @type {Set<Module>***REMOVED*** */
		this.modules = new Set();
		this.modules.add(rootModule);
		/** @type {Map<Module, Problem>***REMOVED*** */
		this.warnings = new Map();
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 */
	add(module) {
		this.modules.add(module);
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @returns {boolean***REMOVED*** true, when the module is in the module set
	 */
	has(module) {
		return this.modules.has(module);
	***REMOVED***

	isEmpty() {
		return this.modules.size === 1;
	***REMOVED***

	/**
	 * @param {Module***REMOVED*** module the module
	 * @param {Problem***REMOVED*** problem the problem
	 */
	addWarning(module, problem) {
		this.warnings.set(module, problem);
	***REMOVED***

	/**
	 * @returns {Map<Module, Problem>***REMOVED*** warnings
	 */
	getWarningsSorted() {
		return new Map(
			Array.from(this.warnings).sort((a, b) => {
				const ai = a[0].identifier();
				const bi = b[0].identifier();
				if (ai < bi) return -1;
				if (ai > bi) return 1;
				return 0;
			***REMOVED***)
		);
	***REMOVED***

	/**
	 * @returns {Set<Module>***REMOVED*** modules as set
	 */
	getModules() {
		return this.modules;
	***REMOVED***

	snapshot() {
		return this.modules.size;
	***REMOVED***

	/**
	 * @param {number***REMOVED*** snapshot snapshot
	 */
	rollback(snapshot) {
		const modules = this.modules;
		for (const m of modules) {
			if (snapshot === 0) {
				modules.delete(m);
			***REMOVED*** else {
				snapshot--;
			***REMOVED***
		***REMOVED***
	***REMOVED***
***REMOVED***

module.exports = ModuleConcatenationPlugin;
