// @babel/plugin-transform-flow-strip-types/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var _helperPluginUtils = require("@babel/helper-plugin-utils");
var _pluginSyntaxFlow = require("@babel/plugin-syntax-flow");
var _core = require("@babel/core");
var _default = exports.default = (0, _helperPluginUtils.declare)((api, opts) => {
  api.assertVersion(7);
  const FLOW_DIRECTIVE = /@flow(?:\s+(?:strict(?:-local)?|weak))?|@noflow/;
  let skipStrip = false;
  const {
    requireDirective = false
  ***REMOVED*** = opts;
  {
    var {
      allowDeclareFields = false
    ***REMOVED*** = opts;
  ***REMOVED***
  return {
    name: "transform-flow-strip-types",
    inherits: _pluginSyntaxFlow.default,
    visitor: {
      Program(path, {
        file: {
          ast: {
            comments
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***) {
        skipStrip = false;
        let directiveFound = false;
        if (comments) {
          for (const comment of comments) {
            if (FLOW_DIRECTIVE.test(comment.value)) {
              directiveFound = true;
              comment.value = comment.value.replace(FLOW_DIRECTIVE, "");
              if (!comment.value.replace(/\*/g, "").trim()) {
                comment.ignore = true;
              ***REMOVED***
            ***REMOVED***
          ***REMOVED***
        ***REMOVED***
        if (!directiveFound && requireDirective) {
          skipStrip = true;
        ***REMOVED***
      ***REMOVED***,
      ImportDeclaration(path) {
        if (skipStrip) return;
        if (!path.node.specifiers.length) return;
        let typeCount = 0;
        path.node.specifiers.forEach(({
          importKind
        ***REMOVED***) => {
          if (importKind === "type" || importKind === "typeof") {
            typeCount++;
          ***REMOVED***
        ***REMOVED***);
        if (typeCount === path.node.specifiers.length) {
          path.remove();
        ***REMOVED***
      ***REMOVED***,
      Flow(path) {
        if (skipStrip) {
          throw path.buildCodeFrameError("A @flow directive is required when using Flow annotations with " + "the `requireDirective` option.");
        ***REMOVED***
        path.remove();
      ***REMOVED***,
      ClassPrivateProperty(path) {
        if (skipStrip) return;
        path.node.typeAnnotation = null;
      ***REMOVED***,
      Class(path) {
        if (skipStrip) return;
        path.node.implements = null;
        path.get("body.body").forEach(child => {
          if (child.isClassProperty()) {
            const {
              node
            ***REMOVED*** = child;
            {
              if (!allowDeclareFields && node.declare) {
                throw child.buildCodeFrameError(`The 'declare' modifier is only allowed when the ` + `'allowDeclareFields' option of ` + `@babel/plugin-transform-flow-strip-types or ` + `@babel/preset-flow is enabled.`);
              ***REMOVED***
            ***REMOVED***
            if (node.declare) {
              child.remove();
            ***REMOVED*** else {
              {
                if (!allowDeclareFields && !node.value && !node.decorators) {
                  child.remove();
                  return;
                ***REMOVED***
              ***REMOVED***
              node.variance = null;
              node.typeAnnotation = null;
            ***REMOVED***
          ***REMOVED***
        ***REMOVED***);
      ***REMOVED***,
      AssignmentPattern({
        node
      ***REMOVED***) {
        if (skipStrip) return;
        if (node.left.optional) {
          node.left.optional = false;
        ***REMOVED***
      ***REMOVED***,
      Function({
        node
      ***REMOVED***) {
        if (skipStrip) return;
        if (node.params.length > 0 && node.params[0].type === "Identifier" && node.params[0].name === "this") {
          node.params.shift();
        ***REMOVED***
        for (let i = 0; i < node.params.length; i++) {
          let param = node.params[i];
          if (param.type === "AssignmentPattern") {
            param = param.left;
          ***REMOVED***
          if (param.optional) {
            param.optional = false;
          ***REMOVED***
        ***REMOVED***
        if (!_core.types.isMethod(node)) {
          node.predicate = null;
        ***REMOVED***
      ***REMOVED***,
      TypeCastExpression(path) {
        if (skipStrip) return;
        let {
          node
        ***REMOVED*** = path;
        do {
          node = node.expression;
        ***REMOVED*** while (_core.types.isTypeCastExpression(node));
        path.replaceWith(node);
      ***REMOVED***,
      CallExpression({
        node
      ***REMOVED***) {
        if (skipStrip) return;
        node.typeArguments = null;
      ***REMOVED***,
      JSXOpeningElement({
        node
      ***REMOVED***) {
        if (skipStrip) return;
        node.typeArguments = null;
      ***REMOVED***,
      OptionalCallExpression({
        node
      ***REMOVED***) {
        if (skipStrip) return;
        node.typeArguments = null;
      ***REMOVED***,
      NewExpression({
        node
      ***REMOVED***) {
        if (skipStrip) return;
        node.typeArguments = null;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;
***REMOVED***);
