// es-iterator-helpers/test/Iterator.js
'use strict';

var defineProperties = require('define-properties');
var test = require('tape');
var functionsHaveNames = require('functions-have-names')();

var index = require('../Iterator');
var impl = require('../Iterator/implementation');

var isEnumerable = Object.prototype.propertyIsEnumerable;

module.exports = {
	tests: function (Iter, name, t) {
		t.equal(typeof Iter, 'function', name + ' is a function');

		t['throws'](
			function () { Iter(); ***REMOVED***, // eslint-disable-line new-cap
			TypeError,
			name + ' throws when Call-ed'
		);

		t['throws'](
			function () { return new Iter(); ***REMOVED***,
			TypeError,
			name + ' throws when Construct-ed'
		);

		var SubIter;
		var SubSubIter;
		try {
			/* eslint no-new-func: 0 */
			SubIter = Function('Iter', 'return class SubIter extends Iter {***REMOVED***;')(Iter);
			SubSubIter = Function('SubIter', 'return class SubSubIter extends SubIter {***REMOVED***;')(SubIter);
		***REMOVED*** catch (e) { /**/ ***REMOVED***

		t.test('class inheritance', { skip: !SubIter ***REMOVED***, function (st) {
			st.doesNotThrow(
				function () { return new SubIter(); ***REMOVED***,
				'Extending ' + name + ' does not throw when Construct-ed'
			);
			st.doesNotThrow(
				function () { return new SubSubIter(); ***REMOVED***,
				'Extending ' + name + ' twice does not throw when Construct-ed'
			);

			st.end();
		***REMOVED***);
	***REMOVED***,
	index: function () {
		test('Iterator: index', function (t) {
			module.exports.tests(index, 'Iterator', t);

			t.end();
		***REMOVED***);
	***REMOVED***,
	implementation: function () {
		test('Iterator: implementation', function (t) {
			module.exports.tests(impl, 'Iterator', t);

			t.end();
		***REMOVED***);
	***REMOVED***,
	shimmed: function () {
		test('Iterator: shimmed', function (t) {
			t.test('Function name', { skip: !functionsHaveNames ***REMOVED***, function (st) {
				st.equal(Iterator.name, 'Iterator', 'Iterator has name "Iterator"');
				st.end();
			***REMOVED***);

			t.test('enumerability', { skip: !defineProperties.supportsDescriptors ***REMOVED***, function (et) {
				et.equal(false, isEnumerable.call(global, Iterator), 'Iterator is not enumerable');
				et.end();
			***REMOVED***);

			t.test('prototype descriptor', { skip: !defineProperties.supportsDescriptors ***REMOVED***, function (pt) {
				var desc = Object.getOwnPropertyDescriptor(Iterator, 'prototype');
				pt.deepEqual(
					desc,
					{
						configurable: false,
						enumerable: false,
						value: Iterator.prototype,
						writable: false
					***REMOVED***
				);

				pt.end();
			***REMOVED***);

			module.exports.tests(Iterator, 'Iterator', t);

			t.end();
		***REMOVED***);
	***REMOVED***
***REMOVED***;
