// es-iterator-helpers/test/Iterator.prototype.filter.js
'use strict';

var defineProperties = require('define-properties');
var test = require('tape');
var callBind = require('call-bind');
var functionsHaveNames = require('functions-have-names')();
var hasStrictMode = require('has-strict-mode')();
var forEach = require('for-each');
var debug = require('object-inspect');
var v = require('es-value-fixtures');
var hasSymbols = require('has-symbols/shams')();
var hasPropertyDescriptors = require('has-property-descriptors')();
var iterate = require('iterate-iterator');

var index = require('../Iterator.prototype.filter');
var impl = require('../Iterator.prototype.filter/implementation');

var fnName = 'filter';

var isEnumerable = Object.prototype.propertyIsEnumerable;

var testIterator = require('./helpers/testIterator');

module.exports = {
	tests: function (filter, name, t) {
		t['throws'](
			function () { return new filter(); ***REMOVED***, // eslint-disable-line new-cap
			TypeError,
			'`' + name + '` is not a constructor'
		);

		forEach(v.primitives.concat(v.objects), function (nonIterator) {
			t['throws'](
				function () { iterate(filter(nonIterator, function () {***REMOVED***)); ***REMOVED***,
				TypeError,
				debug(nonIterator) + ' is not an Object with a callable `next` method'
			);

			var badNext = { next: nonIterator ***REMOVED***;
			t['throws'](
				function () { iterate(filter(badNext, function () {***REMOVED***)); ***REMOVED***,
				TypeError,
				debug(badNext) + ' is not an Object with a callable `next` method'
			);
		***REMOVED***);

		forEach(v.nonFunctions, function (nonFunction) {
			t['throws'](
				function () { filter({ next: function () {***REMOVED*** ***REMOVED***, nonFunction); ***REMOVED***,
				TypeError,
				debug(nonFunction) + ' is not a function'
			);
		***REMOVED***);

		t.test('observable lookups', { skip: !hasPropertyDescriptors ***REMOVED***, function (st) {
			var effects = [];

			var obj = {***REMOVED***;
			Object.defineProperty(obj, 'next', {
				configurable: true,
				enumerable: true,
				get: function next() {
					effects.push('get next');
					return function () {
						return { done: true, value: undefined ***REMOVED***;
					***REMOVED***;
				***REMOVED***
			***REMOVED***);

			st['throws'](
				function () { filter(obj, null); ***REMOVED***,
				TypeError
			);

			st.deepEqual(effects, []);

			st.end();
		***REMOVED***);

		t.test('actual iteration', { skip: !hasSymbols ***REMOVED***, function (st) {
			var arr = [1, 2, 3];
			var iterator = callBind(arr[Symbol.iterator], arr);

			st['throws'](
				function () { return new filter(iterator()); ***REMOVED***, // eslint-disable-line new-cap
				TypeError,
				'`' + name + '` iterator is not a constructor'
			);
			st['throws'](
				function () { return new filter(iterator(), function () {***REMOVED***); ***REMOVED***, // eslint-disable-line new-cap
				TypeError,
				'`' + name + '` iterator is not a constructor'
			);

			testIterator(iterator(), [1, 2, 3], st, 'original');
			testIterator(filter(iterator(), function () { return false; ***REMOVED***), [], st, 'filter for always-false');
			testIterator(filter(iterator(), function () { return true; ***REMOVED***), [1, 2, 3], st, 'filter for always-true');
			testIterator(filter(iterator(), function (x, i) { return x === 2 && i === 1; ***REMOVED***), [2], st, 'filter returns value for matching value/index');

			st.end();
		***REMOVED***);

		t.test('262: test/built-ins/Iterator/prototype/filter/predicate-args', function (st) {
			var g = function g() {
				var arr = ['a', 'b', 'c'];
				var i = 0;
				return {
					next: function () {
						try {
							return {
								value: arr[i],
								done: i >= arr.length
							***REMOVED***;
						***REMOVED*** finally {
							i += 1;
						***REMOVED***
					***REMOVED***
				***REMOVED***;
			***REMOVED***;
			var assertionCount = 0;
			var iter = filter(
				g(),
				function (value, count) {
					if (value === 'a') {
						st.equal(count, 0, 'first iteration');
					***REMOVED*** else if (value === 'b') {
						st.equal(count, 1, 'second iteration');
					***REMOVED*** else if (value === 'c') {
						st.equal(count, 2, 'third iteration');
					***REMOVED*** else {
						st.fail('unexpected iteration');
					***REMOVED***
					assertionCount += 1;
					return true;
				***REMOVED***
			);

			st.equal(assertionCount, 0, 'prior to iteration');

			testIterator(iter, ['a', 'b', 'c'], st, 'iteration');

			st.equal(assertionCount, 3);

			st.end();
		***REMOVED***);

		t.test('262: test/built-ins/Iterator/prototype/filter/predicate-throws', function (st) {
			var returnCalls = 0;

			var iter = {
				next: function () {
					return {
						done: false,
						value: 1
					***REMOVED***;
				***REMOVED***,
				'return': function () {
					returnCalls += 1;
					return {***REMOVED***;
				***REMOVED***
			***REMOVED***;

			var callbackCalls = 0;
			var iterator = filter(iter, function () {
				callbackCalls += 1;
				throw new SyntaxError();
			***REMOVED***);

			st['throws'](function () { iterator.next(); ***REMOVED***, SyntaxError, 'next() throws');

			st.equal(callbackCalls, 1);
			st.equal(returnCalls, 1);

			st.end();
		***REMOVED***);

		t.test('262: test/built-ins/Iterator/prototype/filter/predicate-throws-then-closing-iterator-also-throws', function (st) {
			var iter = {
				next: function next() {
					return {
						done: false,
						value: 1
					***REMOVED***;
				***REMOVED***,
				'return': function () {
					throw new EvalError();
				***REMOVED***
			***REMOVED***;

			var iterator = filter(iter, function () {
				throw new SyntaxError();
			***REMOVED***);

			st['throws'](
				function () { iterator.next(); ***REMOVED***,
				SyntaxError,
				'when the predicate and return() both throw, the predicate’s exception wins'
			);

			st.end();
		***REMOVED***);

		t.test('262: test/built-ins/Iterator/prototype/filter/get-return-method-throws', { skip: !hasPropertyDescriptors ***REMOVED***, function (st) {
			var badIterator = {
				next: function next() {
					return {
						done: false,
						value: 1
					***REMOVED***;
				***REMOVED***
			***REMOVED***;

			Object.defineProperty(badIterator, 'return', { get: function () { throw new SyntaxError(); ***REMOVED*** ***REMOVED***);

			var iter = filter(badIterator, function () { return true; ***REMOVED***);
			iter.next();

			st['throws'](
				function () { iter['return'](); ***REMOVED***,
				SyntaxError,
				'gets the `return` method, whose getter throws'
			);

			st.end();
		***REMOVED***);

		t.test('262: test/built-ins/Iterator/prototype/drop/return-is-forwarded', function (st) {
			var returnCount = 0;

			var badIterator = {
				next: function next() {
					return {
						done: false,
						value: 1
					***REMOVED***;
				***REMOVED***,
				'return': function () {
					returnCount += 1;
					return {***REMOVED***;
				***REMOVED***
			***REMOVED***;

			var iter1 = filter(badIterator, function () { return false; ***REMOVED***);
			st.equal(returnCount, 0, 'iter1, before return()');
			iter1['return']();
			st.equal(returnCount, 1, 'iter1, after return()');

			st.end();
		***REMOVED***);

		t.test('262: test/built-ins/Iterator/prototype/drop/return-is-not-forwarded-after-exhaustion', { skip: !hasPropertyDescriptors ***REMOVED***, function (st) {
			var makeBadIterator = function makeBadIterator() {
				return {
					next: function next() {
						return {
							done: true,
							value: undefined
						***REMOVED***;
					***REMOVED***,
					'return': function () {
						throw new SyntaxError();
					***REMOVED***
				***REMOVED***;
			***REMOVED***;

			var iter1 = filter(makeBadIterator(), function () { return true; ***REMOVED***);
			st['throws'](
				function () { iter1['return'](); ***REMOVED***,
				SyntaxError,
				'iter1, return() throws'
			);
			iter1.next();
			iter1['return']();

			// 3 filters (i wish i had pipeline)
			var iter2 = filter(
				filter(
					filter(
						makeBadIterator(),
						function () { return true; ***REMOVED***
					),
					function () { return true; ***REMOVED***
				),
				function () { return true; ***REMOVED***
			);
			st['throws'](
				function () { iter2['return'](); ***REMOVED***,
				SyntaxError,
				'iter2, return() throws'
			);
			iter2.next();
			iter2['return']();

			st.end();
		***REMOVED***);
	***REMOVED***,
	index: function () {
		test('Iterator.prototype.' + fnName + ': index', function (t) {
			module.exports.tests(index, 'Iterator.prototype.' + fnName, t);

			t.end();
		***REMOVED***);
	***REMOVED***,
	implementation: function () {
		test('Iterator.prototype.' + fnName + ': implementation', function (t) {
			module.exports.tests(callBind(impl), 'Iterator.prototype.' + fnName, t);

			t.end();
		***REMOVED***);
	***REMOVED***,
	shimmed: function () {
		test('Iterator.prototype.' + fnName + ': shimmed', function (t) {
			t.test('Function name', { skip: !functionsHaveNames ***REMOVED***, function (st) {
				st.equal(Iterator.prototype[fnName].name, fnName, 'Iterator#' + fnName + ' has name "' + fnName + '"');
				st.end();
			***REMOVED***);

			t.test('enumerability', { skip: !defineProperties.supportsDescriptors ***REMOVED***, function (et) {
				et.equal(false, isEnumerable.call(Iterator.prototype, fnName), 'Iterator#' + fnName + ' is not enumerable');
				et.end();
			***REMOVED***);

			t.test('bad string/this value', { skip: !hasStrictMode ***REMOVED***, function (st) {
				st['throws'](function () { return Iterator.prototype[fnName].call(undefined, 'a'); ***REMOVED***, TypeError, 'undefined is not an object');
				st['throws'](function () { return Iterator.prototype[fnName].call(null, 'a'); ***REMOVED***, TypeError, 'null is not an object');
				st.end();
			***REMOVED***);

			module.exports.tests(callBind(Iterator.prototype[fnName]), 'Iterator.prototype.' + fnName, t);

			t.end();
		***REMOVED***);
	***REMOVED***
***REMOVED***;
