// es-iterator-helpers/test/tests.js
'use strict';

var data = {
	anchor: { arg: 'bar"baz"', expected: '<a name="bar&quot;baz&quot;">foo</a>' ***REMOVED***,
	big: '<big>foo</big>',
	blink: '<blink>foo</blink>',
	bold: '<b>foo</b>',
	fixed: '<tt>foo</tt>',
	fontcolor: { arg: 'blue"red"green', expected: '<font color="blue&quot;red&quot;green">foo</font>' ***REMOVED***,
	fontsize: { arg: '10"large"small', expected: '<font size="10&quot;large&quot;small">foo</font>' ***REMOVED***,
	italics: '<i>foo</i>',
	link: { arg: 'url"http://"', expected: '<a href="url&quot;http://&quot;">foo</a>' ***REMOVED***,
	small: '<small>foo</small>',
	strike: '<strike>foo</strike>',
	sub: '<sub>foo</sub>',
	sup: '<sup>foo</sup>'
***REMOVED***;

module.exports = function (method, name, t) {
	var result = data[name] || {***REMOVED***;
	var expected = typeof result === 'string' ? result : result.expected;
	var actual = typeof result === 'string' ? method('foo') : method('foo', result.arg);

	t.equal(actual, expected, name + ': got expected result');
***REMOVED***;
