// jsdom/lib/jsdom/living/mutation-observer/MutationRecord-impl.js
"use strict";

const NodeList = require("../generated/NodeList");

// https://dom.spec.whatwg.org/#mutationrecord
class MutationRecordImpl {
  constructor(globalObject, args, privateData) {
    this._globalObject = globalObject;

    this.type = privateData.type;
    this.target = privateData.target;
    this.previousSibling = privateData.previousSibling;
    this.nextSibling = privateData.nextSibling;
    this.attributeName = privateData.attributeName;
    this.attributeNamespace = privateData.attributeNamespace;
    this.oldValue = privateData.oldValue;

    this._addedNodes = privateData.addedNodes;
    this._removedNodes = privateData.removedNodes;
  ***REMOVED***

  get addedNodes() {
    return NodeList.createImpl(this._globalObject, [], {
      nodes: this._addedNodes
    ***REMOVED***);
  ***REMOVED***

  get removedNodes() {
    return NodeList.createImpl(this._globalObject, [], {
      nodes: this._removedNodes
    ***REMOVED***);
  ***REMOVED***
***REMOVED***

module.exports = {
  implementation: MutationRecordImpl
***REMOVED***;
