// pretty-format/build/index.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
***REMOVED***);
exports.default = exports.DEFAULT_OPTIONS = void 0;
exports.format = format;
exports.plugins = void 0;

var _ansiStyles = _interopRequireDefault(require('ansi-styles'));

var _collections = require('./collections');

var _AsymmetricMatcher = _interopRequireDefault(
  require('./plugins/AsymmetricMatcher')
);

var _ConvertAnsi = _interopRequireDefault(require('./plugins/ConvertAnsi'));

var _DOMCollection = _interopRequireDefault(require('./plugins/DOMCollection'));

var _DOMElement = _interopRequireDefault(require('./plugins/DOMElement'));

var _Immutable = _interopRequireDefault(require('./plugins/Immutable'));

var _ReactElement = _interopRequireDefault(require('./plugins/ReactElement'));

var _ReactTestComponent = _interopRequireDefault(
  require('./plugins/ReactTestComponent')
);

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {default: obj***REMOVED***;
***REMOVED***

/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/* eslint-disable local/ban-types-eventually */
const toString = Object.prototype.toString;
const toISOString = Date.prototype.toISOString;
const errorToString = Error.prototype.toString;
const regExpToString = RegExp.prototype.toString;
/**
 * Explicitly comparing typeof constructor to function avoids undefined as name
 * when mock identity-obj-proxy returns the key as the value for any key.
 */

const getConstructorName = val =>
  (typeof val.constructor === 'function' && val.constructor.name) || 'Object';
/* global window */

/** Is val is equal to global window object? Works even if it does not exist :) */

const isWindow = val => typeof window !== 'undefined' && val === window;

const SYMBOL_REGEXP = /^Symbol\((.*)\)(.*)$/;
const NEWLINE_REGEXP = /\n/gi;

class PrettyFormatPluginError extends Error {
  constructor(message, stack) {
    super(message);
    this.stack = stack;
    this.name = this.constructor.name;
  ***REMOVED***
***REMOVED***

function isToStringedArrayType(toStringed) {
  return (
    toStringed === '[object Array]' ||
    toStringed === '[object ArrayBuffer]' ||
    toStringed === '[object DataView]' ||
    toStringed === '[object Float32Array]' ||
    toStringed === '[object Float64Array]' ||
    toStringed === '[object Int8Array]' ||
    toStringed === '[object Int16Array]' ||
    toStringed === '[object Int32Array]' ||
    toStringed === '[object Uint8Array]' ||
    toStringed === '[object Uint8ClampedArray]' ||
    toStringed === '[object Uint16Array]' ||
    toStringed === '[object Uint32Array]'
  );
***REMOVED***

function printNumber(val) {
  return Object.is(val, -0) ? '-0' : String(val);
***REMOVED***

function printBigInt(val) {
  return String(`${val***REMOVED***n`);
***REMOVED***

function printFunction(val, printFunctionName) {
  if (!printFunctionName) {
    return '[Function]';
  ***REMOVED***

  return `[Function ${val.name || 'anonymous'***REMOVED***]`;
***REMOVED***

function printSymbol(val) {
  return String(val).replace(SYMBOL_REGEXP, 'Symbol($1)');
***REMOVED***

function printError(val) {
  return `[${errorToString.call(val)***REMOVED***]`;
***REMOVED***
/**
 * The first port of call for printing an object, handles most of the
 * data-types in JS.
 */

function printBasicValue(val, printFunctionName, escapeRegex, escapeString) {
  if (val === true || val === false) {
    return `${val***REMOVED***`;
  ***REMOVED***

  if (val === undefined) {
    return 'undefined';
  ***REMOVED***

  if (val === null) {
    return 'null';
  ***REMOVED***

  const typeOf = typeof val;

  if (typeOf === 'number') {
    return printNumber(val);
  ***REMOVED***

  if (typeOf === 'bigint') {
    return printBigInt(val);
  ***REMOVED***

  if (typeOf === 'string') {
    if (escapeString) {
      return `"${val.replace(/"|\\/g, '\\$&')***REMOVED***"`;
    ***REMOVED***

    return `"${val***REMOVED***"`;
  ***REMOVED***

  if (typeOf === 'function') {
    return printFunction(val, printFunctionName);
  ***REMOVED***

  if (typeOf === 'symbol') {
    return printSymbol(val);
  ***REMOVED***

  const toStringed = toString.call(val);

  if (toStringed === '[object WeakMap]') {
    return 'WeakMap {***REMOVED***';
  ***REMOVED***

  if (toStringed === '[object WeakSet]') {
    return 'WeakSet {***REMOVED***';
  ***REMOVED***

  if (
    toStringed === '[object Function]' ||
    toStringed === '[object GeneratorFunction]'
  ) {
    return printFunction(val, printFunctionName);
  ***REMOVED***

  if (toStringed === '[object Symbol]') {
    return printSymbol(val);
  ***REMOVED***

  if (toStringed === '[object Date]') {
    return isNaN(+val) ? 'Date { NaN ***REMOVED***' : toISOString.call(val);
  ***REMOVED***

  if (toStringed === '[object Error]') {
    return printError(val);
  ***REMOVED***

  if (toStringed === '[object RegExp]') {
    if (escapeRegex) {
      // https://github.com/benjamingr/RegExp.escape/blob/main/polyfill.js
      return regExpToString.call(val).replace(/[\\^$*+?.()|[\]{***REMOVED***]/g, '\\$&');
    ***REMOVED***

    return regExpToString.call(val);
  ***REMOVED***

  if (val instanceof Error) {
    return printError(val);
  ***REMOVED***

  return null;
***REMOVED***
/**
 * Handles more complex objects ( such as objects with circular references.
 * maps and sets etc )
 */

function printComplexValue(
  val,
  config,
  indentation,
  depth,
  refs,
  hasCalledToJSON
) {
  if (refs.indexOf(val) !== -1) {
    return '[Circular]';
  ***REMOVED***

  refs = refs.slice();
  refs.push(val);
  const hitMaxDepth = ++depth > config.maxDepth;
  const min = config.min;

  if (
    config.callToJSON &&
    !hitMaxDepth &&
    val.toJSON &&
    typeof val.toJSON === 'function' &&
    !hasCalledToJSON
  ) {
    return printer(val.toJSON(), config, indentation, depth, refs, true);
  ***REMOVED***

  const toStringed = toString.call(val);

  if (toStringed === '[object Arguments]') {
    return hitMaxDepth
      ? '[Arguments]'
      : `${min ? '' : 'Arguments '***REMOVED***[${(0, _collections.printListItems)(
          val,
          config,
          indentation,
          depth,
          refs,
          printer
        )***REMOVED***]`;
  ***REMOVED***

  if (isToStringedArrayType(toStringed)) {
    return hitMaxDepth
      ? `[${val.constructor.name***REMOVED***]`
      : `${
          min
            ? ''
            : !config.printBasicPrototype && val.constructor.name === 'Array'
            ? ''
            : `${val.constructor.name***REMOVED*** `
        ***REMOVED***[${(0, _collections.printListItems)(
          val,
          config,
          indentation,
          depth,
          refs,
          printer
        )***REMOVED***]`;
  ***REMOVED***

  if (toStringed === '[object Map]') {
    return hitMaxDepth
      ? '[Map]'
      : `Map {${(0, _collections.printIteratorEntries)(
          val.entries(),
          config,
          indentation,
          depth,
          refs,
          printer,
          ' => '
        )***REMOVED******REMOVED***`;
  ***REMOVED***

  if (toStringed === '[object Set]') {
    return hitMaxDepth
      ? '[Set]'
      : `Set {${(0, _collections.printIteratorValues)(
          val.values(),
          config,
          indentation,
          depth,
          refs,
          printer
        )***REMOVED******REMOVED***`;
  ***REMOVED*** // Avoid failure to serialize global window object in jsdom test environment.
  // For example, not even relevant if window is prop of React element.

  return hitMaxDepth || isWindow(val)
    ? `[${getConstructorName(val)***REMOVED***]`
    : `${
        min
          ? ''
          : !config.printBasicPrototype && getConstructorName(val) === 'Object'
          ? ''
          : `${getConstructorName(val)***REMOVED*** `
      ***REMOVED***{${(0, _collections.printObjectProperties)(
        val,
        config,
        indentation,
        depth,
        refs,
        printer
      )***REMOVED******REMOVED***`;
***REMOVED***

function isNewPlugin(plugin) {
  return plugin.serialize != null;
***REMOVED***

function printPlugin(plugin, val, config, indentation, depth, refs) {
  let printed;

  try {
    printed = isNewPlugin(plugin)
      ? plugin.serialize(val, config, indentation, depth, refs, printer)
      : plugin.print(
          val,
          valChild => printer(valChild, config, indentation, depth, refs),
          str => {
            const indentationNext = indentation + config.indent;
            return (
              indentationNext +
              str.replace(NEWLINE_REGEXP, `\n${indentationNext***REMOVED***`)
            );
          ***REMOVED***,
          {
            edgeSpacing: config.spacingOuter,
            min: config.min,
            spacing: config.spacingInner
          ***REMOVED***,
          config.colors
        );
  ***REMOVED*** catch (error) {
    throw new PrettyFormatPluginError(error.message, error.stack);
  ***REMOVED***

  if (typeof printed !== 'string') {
    throw new Error(
      `pretty-format: Plugin must return type "string" but instead returned "${typeof printed***REMOVED***".`
    );
  ***REMOVED***

  return printed;
***REMOVED***

function findPlugin(plugins, val) {
  for (let p = 0; p < plugins.length; p++) {
    try {
      if (plugins[p].test(val)) {
        return plugins[p];
      ***REMOVED***
    ***REMOVED*** catch (error) {
      throw new PrettyFormatPluginError(error.message, error.stack);
    ***REMOVED***
  ***REMOVED***

  return null;
***REMOVED***

function printer(val, config, indentation, depth, refs, hasCalledToJSON) {
  const plugin = findPlugin(config.plugins, val);

  if (plugin !== null) {
    return printPlugin(plugin, val, config, indentation, depth, refs);
  ***REMOVED***

  const basicResult = printBasicValue(
    val,
    config.printFunctionName,
    config.escapeRegex,
    config.escapeString
  );

  if (basicResult !== null) {
    return basicResult;
  ***REMOVED***

  return printComplexValue(
    val,
    config,
    indentation,
    depth,
    refs,
    hasCalledToJSON
  );
***REMOVED***

const DEFAULT_THEME = {
  comment: 'gray',
  content: 'reset',
  prop: 'yellow',
  tag: 'cyan',
  value: 'green'
***REMOVED***;
const DEFAULT_THEME_KEYS = Object.keys(DEFAULT_THEME);
const DEFAULT_OPTIONS = {
  callToJSON: true,
  compareKeys: undefined,
  escapeRegex: false,
  escapeString: true,
  highlight: false,
  indent: 2,
  maxDepth: Infinity,
  maxWidth: Infinity,
  min: false,
  plugins: [],
  printBasicPrototype: true,
  printFunctionName: true,
  theme: DEFAULT_THEME
***REMOVED***;
exports.DEFAULT_OPTIONS = DEFAULT_OPTIONS;

function validateOptions(options) {
  Object.keys(options).forEach(key => {
    if (!Object.prototype.hasOwnProperty.call(DEFAULT_OPTIONS, key)) {
      throw new Error(`pretty-format: Unknown option "${key***REMOVED***".`);
    ***REMOVED***
  ***REMOVED***);

  if (options.min && options.indent !== undefined && options.indent !== 0) {
    throw new Error(
      'pretty-format: Options "min" and "indent" cannot be used together.'
    );
  ***REMOVED***

  if (options.theme !== undefined) {
    if (options.theme === null) {
      throw new Error('pretty-format: Option "theme" must not be null.');
    ***REMOVED***

    if (typeof options.theme !== 'object') {
      throw new Error(
        `pretty-format: Option "theme" must be of type "object" but instead received "${typeof options.theme***REMOVED***".`
      );
    ***REMOVED***
  ***REMOVED***
***REMOVED***

const getColorsHighlight = options =>
  DEFAULT_THEME_KEYS.reduce((colors, key) => {
    const value =
      options.theme && options.theme[key] !== undefined
        ? options.theme[key]
        : DEFAULT_THEME[key];
    const color = value && _ansiStyles.default[value];

    if (
      color &&
      typeof color.close === 'string' &&
      typeof color.open === 'string'
    ) {
      colors[key] = color;
    ***REMOVED*** else {
      throw new Error(
        `pretty-format: Option "theme" has a key "${key***REMOVED***" whose value "${value***REMOVED***" is undefined in ansi-styles.`
      );
    ***REMOVED***

    return colors;
  ***REMOVED***, Object.create(null));

const getColorsEmpty = () =>
  DEFAULT_THEME_KEYS.reduce((colors, key) => {
    colors[key] = {
      close: '',
      open: ''
    ***REMOVED***;
    return colors;
  ***REMOVED***, Object.create(null));

const getPrintFunctionName = options => {
  var _options$printFunctio;

  return (_options$printFunctio =
    options === null || options === void 0
      ? void 0
      : options.printFunctionName) !== null && _options$printFunctio !== void 0
    ? _options$printFunctio
    : DEFAULT_OPTIONS.printFunctionName;
***REMOVED***;

const getEscapeRegex = options => {
  var _options$escapeRegex;

  return (_options$escapeRegex =
    options === null || options === void 0 ? void 0 : options.escapeRegex) !==
    null && _options$escapeRegex !== void 0
    ? _options$escapeRegex
    : DEFAULT_OPTIONS.escapeRegex;
***REMOVED***;

const getEscapeString = options => {
  var _options$escapeString;

  return (_options$escapeString =
    options === null || options === void 0 ? void 0 : options.escapeString) !==
    null && _options$escapeString !== void 0
    ? _options$escapeString
    : DEFAULT_OPTIONS.escapeString;
***REMOVED***;

const getConfig = options => {
  var _options$callToJSON,
    _options$indent,
    _options$maxDepth,
    _options$maxWidth,
    _options$min,
    _options$plugins,
    _options$printBasicPr;

  return {
    callToJSON:
      (_options$callToJSON =
        options === null || options === void 0
          ? void 0
          : options.callToJSON) !== null && _options$callToJSON !== void 0
        ? _options$callToJSON
        : DEFAULT_OPTIONS.callToJSON,
    colors:
      options !== null && options !== void 0 && options.highlight
        ? getColorsHighlight(options)
        : getColorsEmpty(),
    compareKeys:
      typeof (options === null || options === void 0
        ? void 0
        : options.compareKeys) === 'function'
        ? options.compareKeys
        : DEFAULT_OPTIONS.compareKeys,
    escapeRegex: getEscapeRegex(options),
    escapeString: getEscapeString(options),
    indent:
      options !== null && options !== void 0 && options.min
        ? ''
        : createIndent(
            (_options$indent =
              options === null || options === void 0
                ? void 0
                : options.indent) !== null && _options$indent !== void 0
              ? _options$indent
              : DEFAULT_OPTIONS.indent
          ),
    maxDepth:
      (_options$maxDepth =
        options === null || options === void 0 ? void 0 : options.maxDepth) !==
        null && _options$maxDepth !== void 0
        ? _options$maxDepth
        : DEFAULT_OPTIONS.maxDepth,
    maxWidth:
      (_options$maxWidth =
        options === null || options === void 0 ? void 0 : options.maxWidth) !==
        null && _options$maxWidth !== void 0
        ? _options$maxWidth
        : DEFAULT_OPTIONS.maxWidth,
    min:
      (_options$min =
        options === null || options === void 0 ? void 0 : options.min) !==
        null && _options$min !== void 0
        ? _options$min
        : DEFAULT_OPTIONS.min,
    plugins:
      (_options$plugins =
        options === null || options === void 0 ? void 0 : options.plugins) !==
        null && _options$plugins !== void 0
        ? _options$plugins
        : DEFAULT_OPTIONS.plugins,
    printBasicPrototype:
      (_options$printBasicPr =
        options === null || options === void 0
          ? void 0
          : options.printBasicPrototype) !== null &&
      _options$printBasicPr !== void 0
        ? _options$printBasicPr
        : true,
    printFunctionName: getPrintFunctionName(options),
    spacingInner:
      options !== null && options !== void 0 && options.min ? ' ' : '\n',
    spacingOuter:
      options !== null && options !== void 0 && options.min ? '' : '\n'
  ***REMOVED***;
***REMOVED***;

function createIndent(indent) {
  return new Array(indent + 1).join(' ');
***REMOVED***
/**
 * Returns a presentation string of your `val` object
 * @param val any potential JavaScript object
 * @param options Custom settings
 */

function format(val, options) {
  if (options) {
    validateOptions(options);

    if (options.plugins) {
      const plugin = findPlugin(options.plugins, val);

      if (plugin !== null) {
        return printPlugin(plugin, val, getConfig(options), '', 0, []);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  const basicResult = printBasicValue(
    val,
    getPrintFunctionName(options),
    getEscapeRegex(options),
    getEscapeString(options)
  );

  if (basicResult !== null) {
    return basicResult;
  ***REMOVED***

  return printComplexValue(val, getConfig(options), '', 0, []);
***REMOVED***

const plugins = {
  AsymmetricMatcher: _AsymmetricMatcher.default,
  ConvertAnsi: _ConvertAnsi.default,
  DOMCollection: _DOMCollection.default,
  DOMElement: _DOMElement.default,
  Immutable: _Immutable.default,
  ReactElement: _ReactElement.default,
  ReactTestComponent: _ReactTestComponent.default
***REMOVED***;
exports.plugins = plugins;
var _default = format;
exports.default = _default;
