// @babel/plugin-transform-exponentiation-operator/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var _helperPluginUtils = require("@babel/helper-plugin-utils");
var _default = exports.default = (0, _helperPluginUtils.declare)(api => {
  api.assertVersion("^7.0.0-0 || >8.0.0-alpha <8.0.0-beta");
  const {
    types: t,
    template
  ***REMOVED*** = api;
  function build(left, right) {
    return t.callExpression(t.memberExpression(t.identifier("Math"), t.identifier("pow")), [left, right]);
  ***REMOVED***
  function maybeMemoize(node, scope) {
    if (scope.isStatic(node)) {
      return {
        assign: node,
        ref: t.cloneNode(node)
      ***REMOVED***;
    ***REMOVED***
    if (scope.path.isPattern()) {
      return null;
    ***REMOVED***
    const id = scope.generateUidIdentifierBasedOnNode(node);
    scope.push({
      id
    ***REMOVED***);
    return {
      assign: t.assignmentExpression("=", t.cloneNode(id), node),
      ref: t.cloneNode(id)
    ***REMOVED***;
  ***REMOVED***
  return {
    name: "transform-exponentiation-operator",
    visitor: {
      AssignmentExpression(path) {
        const {
          node,
          scope
        ***REMOVED*** = path;
        if (node.operator !== "**=") return;
        if (t.isMemberExpression(node.left)) {
          let member1;
          let member2;
          const object = maybeMemoize(node.left.object, scope);
          if (!object) {
            path.replaceWith(template.expression.ast`(() => ${path.node***REMOVED***)()`);
            return;
          ***REMOVED***
          const {
            property,
            computed
          ***REMOVED*** = node.left;
          if (computed) {
            const prop = maybeMemoize(property, scope);
            member1 = t.memberExpression(object.assign, prop.assign, true);
            member2 = t.memberExpression(object.ref, prop.ref, true);
          ***REMOVED*** else {
            member1 = t.memberExpression(object.assign, property, false);
            member2 = t.memberExpression(object.ref, t.cloneNode(property), false);
          ***REMOVED***
          path.replaceWith(t.assignmentExpression("=", member1, build(member2, node.right)));
        ***REMOVED*** else {
          path.replaceWith(t.assignmentExpression("=", node.left, build(t.cloneNode(node.left), node.right)));
        ***REMOVED***
      ***REMOVED***,
      BinaryExpression(path) {
        const {
          node
        ***REMOVED*** = path;
        if (node.operator === "**") {
          path.replaceWith(build(node.left, node.right));
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;
***REMOVED***);
