// aria-query/lib/ariaPropsMap.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var _iterationDecorator = _interopRequireDefault(require("./util/iterationDecorator"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e ***REMOVED***; ***REMOVED***
function _slicedToArray(r, e) { return _arrayWithHoles(r) || _iterableToArrayLimit(r, e) || _unsupportedIterableToArray(r, e) || _nonIterableRest(); ***REMOVED***
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); ***REMOVED***
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {***REMOVED***.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; ***REMOVED*** ***REMOVED***
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; ***REMOVED***
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; ***REMOVED*** else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); ***REMOVED*** catch (r) { o = !0, n = r; ***REMOVED*** finally { try { if (!f && null != t.return && (u = t.return(), Object(u) !== u)) return; ***REMOVED*** finally { if (o) throw n; ***REMOVED*** ***REMOVED*** return a; ***REMOVED*** ***REMOVED***
function _arrayWithHoles(r) { if (Array.isArray(r)) return r; ***REMOVED***
var properties = [['aria-activedescendant', {
  'type': 'id'
***REMOVED***], ['aria-atomic', {
  'type': 'boolean'
***REMOVED***], ['aria-autocomplete', {
  'type': 'token',
  'values': ['inline', 'list', 'both', 'none']
***REMOVED***], ['aria-braillelabel', {
  'type': 'string'
***REMOVED***], ['aria-brailleroledescription', {
  'type': 'string'
***REMOVED***], ['aria-busy', {
  'type': 'boolean'
***REMOVED***], ['aria-checked', {
  'type': 'tristate'
***REMOVED***], ['aria-colcount', {
  type: 'integer'
***REMOVED***], ['aria-colindex', {
  type: 'integer'
***REMOVED***], ['aria-colspan', {
  type: 'integer'
***REMOVED***], ['aria-controls', {
  'type': 'idlist'
***REMOVED***], ['aria-current', {
  type: 'token',
  values: ['page', 'step', 'location', 'date', 'time', true, false]
***REMOVED***], ['aria-describedby', {
  'type': 'idlist'
***REMOVED***], ['aria-description', {
  'type': 'string'
***REMOVED***], ['aria-details', {
  'type': 'id'
***REMOVED***], ['aria-disabled', {
  'type': 'boolean'
***REMOVED***], ['aria-dropeffect', {
  'type': 'tokenlist',
  'values': ['copy', 'execute', 'link', 'move', 'none', 'popup']
***REMOVED***], ['aria-errormessage', {
  'type': 'id'
***REMOVED***], ['aria-expanded', {
  'type': 'boolean',
  'allowundefined': true
***REMOVED***], ['aria-flowto', {
  'type': 'idlist'
***REMOVED***], ['aria-grabbed', {
  'type': 'boolean',
  'allowundefined': true
***REMOVED***], ['aria-haspopup', {
  'type': 'token',
  'values': [false, true, 'menu', 'listbox', 'tree', 'grid', 'dialog']
***REMOVED***], ['aria-hidden', {
  'type': 'boolean',
  'allowundefined': true
***REMOVED***], ['aria-invalid', {
  'type': 'token',
  'values': ['grammar', false, 'spelling', true]
***REMOVED***], ['aria-keyshortcuts', {
  type: 'string'
***REMOVED***], ['aria-label', {
  'type': 'string'
***REMOVED***], ['aria-labelledby', {
  'type': 'idlist'
***REMOVED***], ['aria-level', {
  'type': 'integer'
***REMOVED***], ['aria-live', {
  'type': 'token',
  'values': ['assertive', 'off', 'polite']
***REMOVED***], ['aria-modal', {
  type: 'boolean'
***REMOVED***], ['aria-multiline', {
  'type': 'boolean'
***REMOVED***], ['aria-multiselectable', {
  'type': 'boolean'
***REMOVED***], ['aria-orientation', {
  'type': 'token',
  'values': ['vertical', 'undefined', 'horizontal']
***REMOVED***], ['aria-owns', {
  'type': 'idlist'
***REMOVED***], ['aria-placeholder', {
  type: 'string'
***REMOVED***], ['aria-posinset', {
  'type': 'integer'
***REMOVED***], ['aria-pressed', {
  'type': 'tristate'
***REMOVED***], ['aria-readonly', {
  'type': 'boolean'
***REMOVED***], ['aria-relevant', {
  'type': 'tokenlist',
  'values': ['additions', 'all', 'removals', 'text']
***REMOVED***], ['aria-required', {
  'type': 'boolean'
***REMOVED***], ['aria-roledescription', {
  type: 'string'
***REMOVED***], ['aria-rowcount', {
  type: 'integer'
***REMOVED***], ['aria-rowindex', {
  type: 'integer'
***REMOVED***], ['aria-rowspan', {
  type: 'integer'
***REMOVED***], ['aria-selected', {
  'type': 'boolean',
  'allowundefined': true
***REMOVED***], ['aria-setsize', {
  'type': 'integer'
***REMOVED***], ['aria-sort', {
  'type': 'token',
  'values': ['ascending', 'descending', 'none', 'other']
***REMOVED***], ['aria-valuemax', {
  'type': 'number'
***REMOVED***], ['aria-valuemin', {
  'type': 'number'
***REMOVED***], ['aria-valuenow', {
  'type': 'number'
***REMOVED***], ['aria-valuetext', {
  'type': 'string'
***REMOVED***]];
var ariaPropsMap = {
  entries: function entries() {
    return properties;
  ***REMOVED***,
  forEach: function forEach(fn) {
    var thisArg = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
    for (var _i = 0, _properties = properties; _i < _properties.length; _i++) {
      var _properties$_i = _slicedToArray(_properties[_i], 2),
        key = _properties$_i[0],
        values = _properties$_i[1];
      fn.call(thisArg, values, key, properties);
    ***REMOVED***
  ***REMOVED***,
  get: function get(key) {
    var item = properties.filter(function (tuple) {
      return tuple[0] === key ? true : false;
    ***REMOVED***)[0];
    return item && item[1];
  ***REMOVED***,
  has: function has(key) {
    return !!ariaPropsMap.get(key);
  ***REMOVED***,
  keys: function keys() {
    return properties.map(function (_ref) {
      var _ref2 = _slicedToArray(_ref, 1),
        key = _ref2[0];
      return key;
    ***REMOVED***);
  ***REMOVED***,
  values: function values() {
    return properties.map(function (_ref3) {
      var _ref4 = _slicedToArray(_ref3, 2),
        values = _ref4[1];
      return values;
    ***REMOVED***);
  ***REMOVED***
***REMOVED***;
var _default = exports.default = (0, _iterationDecorator.default)(ariaPropsMap, ariaPropsMap.entries());
