// aria-query/lib/domMap.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var _iterationDecorator = _interopRequireDefault(require("./util/iterationDecorator"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e ***REMOVED***; ***REMOVED***
function _slicedToArray(r, e) { return _arrayWithHoles(r) || _iterableToArrayLimit(r, e) || _unsupportedIterableToArray(r, e) || _nonIterableRest(); ***REMOVED***
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); ***REMOVED***
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {***REMOVED***.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; ***REMOVED*** ***REMOVED***
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; ***REMOVED***
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; ***REMOVED*** else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); ***REMOVED*** catch (r) { o = !0, n = r; ***REMOVED*** finally { try { if (!f && null != t.return && (u = t.return(), Object(u) !== u)) return; ***REMOVED*** finally { if (o) throw n; ***REMOVED*** ***REMOVED*** return a; ***REMOVED*** ***REMOVED***
function _arrayWithHoles(r) { if (Array.isArray(r)) return r; ***REMOVED***
var dom = [['a', {
  reserved: false
***REMOVED***], ['abbr', {
  reserved: false
***REMOVED***], ['acronym', {
  reserved: false
***REMOVED***], ['address', {
  reserved: false
***REMOVED***], ['applet', {
  reserved: false
***REMOVED***], ['area', {
  reserved: false
***REMOVED***], ['article', {
  reserved: false
***REMOVED***], ['aside', {
  reserved: false
***REMOVED***], ['audio', {
  reserved: false
***REMOVED***], ['b', {
  reserved: false
***REMOVED***], ['base', {
  reserved: true
***REMOVED***], ['bdi', {
  reserved: false
***REMOVED***], ['bdo', {
  reserved: false
***REMOVED***], ['big', {
  reserved: false
***REMOVED***], ['blink', {
  reserved: false
***REMOVED***], ['blockquote', {
  reserved: false
***REMOVED***], ['body', {
  reserved: false
***REMOVED***], ['br', {
  reserved: false
***REMOVED***], ['button', {
  reserved: false
***REMOVED***], ['canvas', {
  reserved: false
***REMOVED***], ['caption', {
  reserved: false
***REMOVED***], ['center', {
  reserved: false
***REMOVED***], ['cite', {
  reserved: false
***REMOVED***], ['code', {
  reserved: false
***REMOVED***], ['col', {
  reserved: true
***REMOVED***], ['colgroup', {
  reserved: true
***REMOVED***], ['content', {
  reserved: false
***REMOVED***], ['data', {
  reserved: false
***REMOVED***], ['datalist', {
  reserved: false
***REMOVED***], ['dd', {
  reserved: false
***REMOVED***], ['del', {
  reserved: false
***REMOVED***], ['details', {
  reserved: false
***REMOVED***], ['dfn', {
  reserved: false
***REMOVED***], ['dialog', {
  reserved: false
***REMOVED***], ['dir', {
  reserved: false
***REMOVED***], ['div', {
  reserved: false
***REMOVED***], ['dl', {
  reserved: false
***REMOVED***], ['dt', {
  reserved: false
***REMOVED***], ['em', {
  reserved: false
***REMOVED***], ['embed', {
  reserved: false
***REMOVED***], ['fieldset', {
  reserved: false
***REMOVED***], ['figcaption', {
  reserved: false
***REMOVED***], ['figure', {
  reserved: false
***REMOVED***], ['font', {
  reserved: false
***REMOVED***], ['footer', {
  reserved: false
***REMOVED***], ['form', {
  reserved: false
***REMOVED***], ['frame', {
  reserved: false
***REMOVED***], ['frameset', {
  reserved: false
***REMOVED***], ['h1', {
  reserved: false
***REMOVED***], ['h2', {
  reserved: false
***REMOVED***], ['h3', {
  reserved: false
***REMOVED***], ['h4', {
  reserved: false
***REMOVED***], ['h5', {
  reserved: false
***REMOVED***], ['h6', {
  reserved: false
***REMOVED***], ['head', {
  reserved: true
***REMOVED***], ['header', {
  reserved: false
***REMOVED***], ['hgroup', {
  reserved: false
***REMOVED***], ['hr', {
  reserved: false
***REMOVED***], ['html', {
  reserved: true
***REMOVED***], ['i', {
  reserved: false
***REMOVED***], ['iframe', {
  reserved: false
***REMOVED***], ['img', {
  reserved: false
***REMOVED***], ['input', {
  reserved: false
***REMOVED***], ['ins', {
  reserved: false
***REMOVED***], ['kbd', {
  reserved: false
***REMOVED***], ['keygen', {
  reserved: false
***REMOVED***], ['label', {
  reserved: false
***REMOVED***], ['legend', {
  reserved: false
***REMOVED***], ['li', {
  reserved: false
***REMOVED***], ['link', {
  reserved: true
***REMOVED***], ['main', {
  reserved: false
***REMOVED***], ['map', {
  reserved: false
***REMOVED***], ['mark', {
  reserved: false
***REMOVED***], ['marquee', {
  reserved: false
***REMOVED***], ['menu', {
  reserved: false
***REMOVED***], ['menuitem', {
  reserved: false
***REMOVED***], ['meta', {
  reserved: true
***REMOVED***], ['meter', {
  reserved: false
***REMOVED***], ['nav', {
  reserved: false
***REMOVED***], ['noembed', {
  reserved: true
***REMOVED***], ['noscript', {
  reserved: true
***REMOVED***], ['object', {
  reserved: false
***REMOVED***], ['ol', {
  reserved: false
***REMOVED***], ['optgroup', {
  reserved: false
***REMOVED***], ['option', {
  reserved: false
***REMOVED***], ['output', {
  reserved: false
***REMOVED***], ['p', {
  reserved: false
***REMOVED***], ['param', {
  reserved: true
***REMOVED***], ['picture', {
  reserved: true
***REMOVED***], ['pre', {
  reserved: false
***REMOVED***], ['progress', {
  reserved: false
***REMOVED***], ['q', {
  reserved: false
***REMOVED***], ['rp', {
  reserved: false
***REMOVED***], ['rt', {
  reserved: false
***REMOVED***], ['rtc', {
  reserved: false
***REMOVED***], ['ruby', {
  reserved: false
***REMOVED***], ['s', {
  reserved: false
***REMOVED***], ['samp', {
  reserved: false
***REMOVED***], ['script', {
  reserved: true
***REMOVED***], ['section', {
  reserved: false
***REMOVED***], ['select', {
  reserved: false
***REMOVED***], ['small', {
  reserved: false
***REMOVED***], ['source', {
  reserved: true
***REMOVED***], ['spacer', {
  reserved: false
***REMOVED***], ['span', {
  reserved: false
***REMOVED***], ['strike', {
  reserved: false
***REMOVED***], ['strong', {
  reserved: false
***REMOVED***], ['style', {
  reserved: true
***REMOVED***], ['sub', {
  reserved: false
***REMOVED***], ['summary', {
  reserved: false
***REMOVED***], ['sup', {
  reserved: false
***REMOVED***], ['table', {
  reserved: false
***REMOVED***], ['tbody', {
  reserved: false
***REMOVED***], ['td', {
  reserved: false
***REMOVED***], ['textarea', {
  reserved: false
***REMOVED***], ['tfoot', {
  reserved: false
***REMOVED***], ['th', {
  reserved: false
***REMOVED***], ['thead', {
  reserved: false
***REMOVED***], ['time', {
  reserved: false
***REMOVED***], ['title', {
  reserved: true
***REMOVED***], ['tr', {
  reserved: false
***REMOVED***], ['track', {
  reserved: true
***REMOVED***], ['tt', {
  reserved: false
***REMOVED***], ['u', {
  reserved: false
***REMOVED***], ['ul', {
  reserved: false
***REMOVED***], ['var', {
  reserved: false
***REMOVED***], ['video', {
  reserved: false
***REMOVED***], ['wbr', {
  reserved: false
***REMOVED***], ['xmp', {
  reserved: false
***REMOVED***]];
var domMap = {
  entries: function entries() {
    return dom;
  ***REMOVED***,
  forEach: function forEach(fn) {
    var thisArg = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
    for (var _i = 0, _dom = dom; _i < _dom.length; _i++) {
      var _dom$_i = _slicedToArray(_dom[_i], 2),
        key = _dom$_i[0],
        values = _dom$_i[1];
      fn.call(thisArg, values, key, dom);
    ***REMOVED***
  ***REMOVED***,
  get: function get(key) {
    var item = dom.filter(function (tuple) {
      return tuple[0] === key ? true : false;
    ***REMOVED***)[0];
    return item && item[1];
  ***REMOVED***,
  has: function has(key) {
    return !!domMap.get(key);
  ***REMOVED***,
  keys: function keys() {
    return dom.map(function (_ref) {
      var _ref2 = _slicedToArray(_ref, 1),
        key = _ref2[0];
      return key;
    ***REMOVED***);
  ***REMOVED***,
  values: function values() {
    return dom.map(function (_ref3) {
      var _ref4 = _slicedToArray(_ref3, 2),
        values = _ref4[1];
      return values;
    ***REMOVED***);
  ***REMOVED***
***REMOVED***;
var _default = exports.default = (0, _iterationDecorator.default)(domMap, domMap.entries());
