// sockjs/lib/chunking-test.js
// Generated by CoffeeScript 1.12.7
(function() {
  var utils;

  utils = require('./utils');

  exports.app = {
    chunking_test: function(req, res, _, next_filter) {
      var write;
      res.setHeader('Content-Type', 'application/javascript; charset=UTF-8');
      res.writeHead(200);
      write = (function(_this) {
        return function(payload) {
          var x;
          try {
            return res.write(payload + '\n');
          ***REMOVED*** catch (error) {
            x = error;
          ***REMOVED***
        ***REMOVED***;
      ***REMOVED***)(this);
      utils.timeout_chain([
        [
          0, (function(_this) {
            return function() {
              return write('h');
            ***REMOVED***;
          ***REMOVED***)(this)
        ], [
          1, (function(_this) {
            return function() {
              return write(Array(2049).join(' ') + 'h');
            ***REMOVED***;
          ***REMOVED***)(this)
        ], [
          5, (function(_this) {
            return function() {
              return write('h');
            ***REMOVED***;
          ***REMOVED***)(this)
        ], [
          25, (function(_this) {
            return function() {
              return write('h');
            ***REMOVED***;
          ***REMOVED***)(this)
        ], [
          125, (function(_this) {
            return function() {
              return write('h');
            ***REMOVED***;
          ***REMOVED***)(this)
        ], [
          625, (function(_this) {
            return function() {
              return write('h');
            ***REMOVED***;
          ***REMOVED***)(this)
        ], [
          3125, (function(_this) {
            return function() {
              write('h');
              return res.end();
            ***REMOVED***;
          ***REMOVED***)(this)
        ]
      ]);
      return true;
    ***REMOVED***,
    info: function(req, res, _) {
      var info;
      info = {
        websocket: this.options.websocket,
        origins: !this.options.disable_cors ? ['*:*'] : void 0,
        cookie_needed: !!this.options.jsessionid,
        entropy: utils.random32()
      ***REMOVED***;
      if (typeof this.options.base_url === 'function') {
        info.base_url = this.options.base_url();
      ***REMOVED*** else if (this.options.base_url) {
        info.base_url = this.options.base_url;
      ***REMOVED***
      res.setHeader('Content-Type', 'application/json; charset=UTF-8');
      res.writeHead(200);
      return res.end(JSON.stringify(info));
    ***REMOVED***,
    info_options: function(req, res) {
      res.statusCode = 204;
      res.setHeader('Access-Control-Allow-Methods', 'OPTIONS, GET');
      res.setHeader('Access-Control-Max-Age', res.cache_for);
      return '';
    ***REMOVED***
  ***REMOVED***;

***REMOVED***).call(this);
