// sockjs/lib/transport.js
// Generated by CoffeeScript 1.12.7
(function() {
  var GenericReceiver, MAP, ResponseReceiver, Session, SockJSConnection, Transport, closeFrame, register, stream, utils, uuidv4,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; ***REMOVED*** function ctor() { this.constructor = child; ***REMOVED*** ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; ***REMOVED***,
    hasProp = {***REMOVED***.hasOwnProperty;

  stream = require('stream');

  uuidv4 = require('uuid').v4;

  utils = require('./utils');

  Transport = (function() {
    function Transport() {***REMOVED***

    return Transport;

  ***REMOVED***)();

  Transport.CONNECTING = 0;

  Transport.OPEN = 1;

  Transport.CLOSING = 2;

  Transport.CLOSED = 3;

  closeFrame = function(status, reason) {
    return 'c' + JSON.stringify([status, reason]);
  ***REMOVED***;

  SockJSConnection = (function(superClass) {
    extend(SockJSConnection, superClass);

    function SockJSConnection(_session) {
      this._session = _session;
      this.id = uuidv4();
      this.headers = {***REMOVED***;
      this.prefix = this._session.prefix;
    ***REMOVED***

    SockJSConnection.prototype.toString = function() {
      return '<SockJSConnection ' + this.id + '>';
    ***REMOVED***;

    SockJSConnection.prototype.write = function(string) {
      return this._session.send('' + string);
    ***REMOVED***;

    SockJSConnection.prototype.end = function(string) {
      if (string) {
        this.write(string);
      ***REMOVED***
      this.close();
      return null;
    ***REMOVED***;

    SockJSConnection.prototype.close = function(code, reason) {
      return this._session.close(code, reason);
    ***REMOVED***;

    SockJSConnection.prototype.destroy = function() {
      this.end();
      return this.removeAllListeners();
    ***REMOVED***;

    SockJSConnection.prototype.destroySoon = function() {
      return this.destroy();
    ***REMOVED***;

    return SockJSConnection;

  ***REMOVED***)(stream.Stream);

  SockJSConnection.prototype.__defineGetter__('readable', function() {
    return this._session.readyState === Transport.OPEN;
  ***REMOVED***);

  SockJSConnection.prototype.__defineGetter__('writable', function() {
    return this._session.readyState === Transport.OPEN;
  ***REMOVED***);

  SockJSConnection.prototype.__defineGetter__('readyState', function() {
    return this._session.readyState;
  ***REMOVED***);

  MAP = {***REMOVED***;

  Session = (function() {
    function Session(session_id1, server) {
      this.session_id = session_id1;
      this.heartbeat_delay = server.options.heartbeat_delay;
      this.disconnect_delay = server.options.disconnect_delay;
      this.prefix = server.options.prefix;
      this.send_buffer = [];
      this.is_closing = false;
      this.readyState = Transport.CONNECTING;
      if (this.session_id) {
        MAP[this.session_id] = this;
      ***REMOVED***
      this.timeout_cb = (function(_this) {
        return function() {
          return _this.didTimeout();
        ***REMOVED***;
      ***REMOVED***)(this);
      this.to_tref = setTimeout(this.timeout_cb, this.disconnect_delay);
      this.connection = new SockJSConnection(this);
      this.emit_open = (function(_this) {
        return function() {
          _this.emit_open = null;
          return server.emit('connection', _this.connection);
        ***REMOVED***;
      ***REMOVED***)(this);
    ***REMOVED***

    Session.prototype.register = function(req, recv) {
      if (this.recv) {
        recv.doSendFrame(closeFrame(2010, "Another connection still open"));
        recv.didClose();
        return;
      ***REMOVED***
      if (this.to_tref) {
        clearTimeout(this.to_tref);
        this.to_tref = null;
      ***REMOVED***
      if (this.readyState === Transport.CLOSING) {
        this.flushToRecv(recv);
        recv.doSendFrame(this.close_frame);
        recv.didClose();
        this.to_tref = setTimeout(this.timeout_cb, this.disconnect_delay);
        return;
      ***REMOVED***
      this.recv = recv;
      this.recv.session = this;
      this.decorateConnection(req);
      if (this.readyState === Transport.CONNECTING) {
        this.recv.doSendFrame('o');
        this.readyState = Transport.OPEN;
        process.nextTick(this.emit_open);
      ***REMOVED***
      if (!this.recv) {
        return;
      ***REMOVED***
      this.tryFlush();
    ***REMOVED***;

    Session.prototype.decorateConnection = function(req) {
      var address, headers, i, key, len, ref, remoteAddress, remotePort, socket, x;
      if (!(socket = this.recv.connection)) {
        socket = this.recv.response.connection;
      ***REMOVED***
      try {
        remoteAddress = socket.remoteAddress;
        remotePort = socket.remotePort;
        address = socket.address();
      ***REMOVED*** catch (error) {
        x = error;
      ***REMOVED***
      if (remoteAddress) {
        this.connection.remoteAddress = remoteAddress;
        this.connection.remotePort = remotePort;
        this.connection.address = address;
      ***REMOVED***
      this.connection.url = req.url;
      this.connection.pathname = req.pathname;
      this.connection.protocol = this.recv.protocol;
      headers = {***REMOVED***;
      ref = ['referer', 'x-client-ip', 'x-forwarded-for', 'x-forwarded-host', 'x-forwarded-port', 'x-cluster-client-ip', 'via', 'x-real-ip', 'x-forwarded-proto', 'x-ssl', 'dnt', 'host', 'user-agent', 'accept-language'];
      for (i = 0, len = ref.length; i < len; i++) {
        key = ref[i];
        if (req.headers[key]) {
          headers[key] = req.headers[key];
        ***REMOVED***
      ***REMOVED***
      if (headers) {
        return this.connection.headers = headers;
      ***REMOVED***
    ***REMOVED***;

    Session.prototype.unregister = function() {
      var delay;
      delay = this.recv.delay_disconnect;
      this.recv.session = null;
      this.recv = null;
      if (this.to_tref) {
        clearTimeout(this.to_tref);
      ***REMOVED***
      if (delay) {
        return this.to_tref = setTimeout(this.timeout_cb, this.disconnect_delay);
      ***REMOVED*** else {
        return this.timeout_cb();
      ***REMOVED***
    ***REMOVED***;

    Session.prototype.flushToRecv = function(recv) {
      var ref, sb;
      if (this.send_buffer.length > 0) {
        ref = [this.send_buffer, []], sb = ref[0], this.send_buffer = ref[1];
        recv.doSendBulk(sb);
        return true;
      ***REMOVED***
      return false;
    ***REMOVED***;

    Session.prototype.tryFlush = function() {
      var x;
      if (!this.flushToRecv(this.recv) || !this.to_tref) {
        if (this.to_tref) {
          clearTimeout(this.to_tref);
        ***REMOVED***
        x = (function(_this) {
          return function() {
            if (_this.recv) {
              _this.to_tref = setTimeout(x, _this.heartbeat_delay);
              return _this.recv.heartbeat();
            ***REMOVED***
          ***REMOVED***;
        ***REMOVED***)(this);
        this.to_tref = setTimeout(x, this.heartbeat_delay);
      ***REMOVED***
    ***REMOVED***;

    Session.prototype.didTimeout = function() {
      if (this.to_tref) {
        clearTimeout(this.to_tref);
        this.to_tref = null;
      ***REMOVED***
      if (this.readyState !== Transport.CONNECTING && this.readyState !== Transport.OPEN && this.readyState !== Transport.CLOSING) {
        throw Error('INVALID_STATE_ERR');
      ***REMOVED***
      if (this.recv) {
        throw Error('RECV_STILL_THERE');
      ***REMOVED***
      this.readyState = Transport.CLOSED;
      this.connection.emit('end');
      this.connection.emit('close');
      this.connection = null;
      if (this.session_id) {
        delete MAP[this.session_id];
        return this.session_id = null;
      ***REMOVED***
    ***REMOVED***;

    Session.prototype.didMessage = function(payload) {
      if (this.readyState === Transport.OPEN) {
        this.connection.emit('data', payload);
      ***REMOVED***
    ***REMOVED***;

    Session.prototype.send = function(payload) {
      if (this.readyState !== Transport.OPEN) {
        return false;
      ***REMOVED***
      this.send_buffer.push('' + payload);
      if (this.recv) {
        this.tryFlush();
      ***REMOVED***
      return true;
    ***REMOVED***;

    Session.prototype.close = function(status, reason) {
      if (status == null) {
        status = 1000;
      ***REMOVED***
      if (reason == null) {
        reason = "Normal closure";
      ***REMOVED***
      if (this.readyState !== Transport.OPEN) {
        return false;
      ***REMOVED***
      this.readyState = Transport.CLOSING;
      this.close_frame = closeFrame(status, reason);
      if (this.recv) {
        this.recv.doSendFrame(this.close_frame);
        if (this.recv) {
          this.recv.didClose();
        ***REMOVED***
        if (this.recv) {
          this.unregister();
        ***REMOVED***
      ***REMOVED***
      return true;
    ***REMOVED***;

    return Session;

  ***REMOVED***)();

  Session.bySessionId = function(session_id) {
    if (!session_id) {
      return null;
    ***REMOVED***
    return MAP[session_id] || null;
  ***REMOVED***;

  register = function(req, server, session_id, receiver) {
    var session;
    session = Session.bySessionId(session_id);
    if (!session) {
      session = new Session(session_id, server);
    ***REMOVED***
    session.register(req, receiver);
    return session;
  ***REMOVED***;

  exports.register = function(req, server, receiver) {
    return register(req, server, req.session, receiver);
  ***REMOVED***;

  exports.registerNoSession = function(req, server, receiver) {
    return register(req, server, void 0, receiver);
  ***REMOVED***;

  GenericReceiver = (function() {
    function GenericReceiver(thingy) {
      this.thingy = thingy;
      this.setUp(this.thingy);
    ***REMOVED***

    GenericReceiver.prototype.setUp = function() {
      this.thingy_end_cb = (function(_this) {
        return function() {
          return _this.didAbort();
        ***REMOVED***;
      ***REMOVED***)(this);
      this.thingy.addListener('close', this.thingy_end_cb);
      return this.thingy.addListener('end', this.thingy_end_cb);
    ***REMOVED***;

    GenericReceiver.prototype.tearDown = function() {
      this.thingy.removeListener('close', this.thingy_end_cb);
      this.thingy.removeListener('end', this.thingy_end_cb);
      return this.thingy_end_cb = null;
    ***REMOVED***;

    GenericReceiver.prototype.didAbort = function() {
      this.delay_disconnect = false;
      return this.didClose();
    ***REMOVED***;

    GenericReceiver.prototype.didClose = function() {
      if (this.thingy) {
        this.tearDown(this.thingy);
        this.thingy = null;
      ***REMOVED***
      if (this.session) {
        return this.session.unregister();
      ***REMOVED***
    ***REMOVED***;

    GenericReceiver.prototype.doSendBulk = function(messages) {
      var m, q_msgs;
      q_msgs = (function() {
        var i, len, results;
        results = [];
        for (i = 0, len = messages.length; i < len; i++) {
          m = messages[i];
          results.push(utils.quote(m));
        ***REMOVED***
        return results;
      ***REMOVED***)();
      return this.doSendFrame('a' + '[' + q_msgs.join(',') + ']');
    ***REMOVED***;

    GenericReceiver.prototype.heartbeat = function() {
      return this.doSendFrame('h');
    ***REMOVED***;

    return GenericReceiver;

  ***REMOVED***)();

  ResponseReceiver = (function(superClass) {
    extend(ResponseReceiver, superClass);

    ResponseReceiver.prototype.max_response_size = void 0;

    ResponseReceiver.prototype.delay_disconnect = true;

    function ResponseReceiver(request, response, options) {
      var x;
      this.request = request;
      this.response = response;
      this.options = options;
      this.curr_response_size = 0;
      try {
        this.request.connection.setKeepAlive(true, 5000);
      ***REMOVED*** catch (error) {
        x = error;
      ***REMOVED***
      ResponseReceiver.__super__.constructor.call(this, this.request.connection);
      if (this.max_response_size === void 0) {
        this.max_response_size = this.options.response_limit;
      ***REMOVED***
    ***REMOVED***

    ResponseReceiver.prototype.doSendFrame = function(payload) {
      var r, x;
      this.curr_response_size += payload.length;
      r = false;
      try {
        this.response.write(payload);
        r = true;
      ***REMOVED*** catch (error) {
        x = error;
      ***REMOVED***
      if (this.max_response_size && this.curr_response_size >= this.max_response_size) {
        this.didClose();
      ***REMOVED***
      return r;
    ***REMOVED***;

    ResponseReceiver.prototype.didClose = function() {
      var x;
      ResponseReceiver.__super__.didClose.apply(this, arguments);
      try {
        this.response.end();
      ***REMOVED*** catch (error) {
        x = error;
      ***REMOVED***
      return this.response = null;
    ***REMOVED***;

    return ResponseReceiver;

  ***REMOVED***)(GenericReceiver);

  exports.GenericReceiver = GenericReceiver;

  exports.Transport = Transport;

  exports.Session = Session;

  exports.ResponseReceiver = ResponseReceiver;

  exports.SockJSConnection = SockJSConnection;

***REMOVED***).call(this);
