// postcss-minify-params/src/index.js
'use strict';
const browserslist = require('browserslist');
const valueParser = require('postcss-value-parser');
const { getArguments ***REMOVED*** = require('cssnano-utils');

/**
 * Return the greatest common divisor
 * of two numbers.
 *
 * @param {number***REMOVED*** a
 * @param {number***REMOVED*** b
 * @return {number***REMOVED***
 */
function gcd(a, b) {
  return b ? gcd(b, a % b) : a;
***REMOVED***

/**
 * @param {number***REMOVED*** a
 * @param {number***REMOVED*** b
 * @return {[number, number]***REMOVED***
 */
function aspectRatio(a, b) {
  const divisor = gcd(a, b);

  return [a / divisor, b / divisor];
***REMOVED***

/**
 * @param {valueParser.Node[]***REMOVED*** args
 * @return {string***REMOVED***
 */
function split(args) {
  return args.map((arg) => valueParser.stringify(arg)).join('');
***REMOVED***

/**
 * @param {valueParser.Node***REMOVED*** node
 * @return {void***REMOVED***
 */
function removeNode(node) {
  node.value = '';
  node.type = 'word';
***REMOVED***

/**
 * @param {unknown[]***REMOVED*** items
 * @return {string***REMOVED***
 */
function sortAndDedupe(items) {
  const a = [...new Set(items)];
  a.sort();
  return a.join();
***REMOVED***

/**
 * @param {boolean***REMOVED*** legacy
 * @param {import('postcss').AtRule***REMOVED*** rule
 * @return {void***REMOVED***
 */
function transform(legacy, rule) {
  const ruleName = rule.name.toLowerCase();

  // We should re-arrange parameters only for `@media` and `@supports` at-rules
  if (!rule.params || !['media', 'supports'].includes(ruleName)) {
    return;
  ***REMOVED***

  const params = valueParser(rule.params);

  params.walk((node, index) => {
    if (node.type === 'div') {
      node.before = node.after = '';
    ***REMOVED*** else if (node.type === 'function') {
      node.before = '';
      if (
        node.nodes[0] &&
        node.nodes[0].type === 'word' &&
        node.nodes[0].value.startsWith('--') &&
        node.nodes[2] === undefined
      ) {
        node.after = ' ';
      ***REMOVED*** else {
        node.after = '';
      ***REMOVED***
      if (
        node.nodes[4] &&
        node.nodes[0].value.toLowerCase().indexOf('-aspect-ratio') === 3
      ) {
        const [a, b] = aspectRatio(
          Number(node.nodes[2].value),
          Number(node.nodes[4].value)
        );

        node.nodes[2].value = a.toString();
        node.nodes[4].value = b.toString();
      ***REMOVED***
    ***REMOVED*** else if (node.type === 'space') {
      node.value = ' ';
    ***REMOVED*** else {
      const prevWord = params.nodes[index - 2];

      if (
        node.value.toLowerCase() === 'all' &&
        rule.name.toLowerCase() === 'media' &&
        !prevWord
      ) {
        const nextWord = params.nodes[index + 2];

        if (!legacy || nextWord) {
          removeNode(node);
        ***REMOVED***

        if (nextWord && nextWord.value.toLowerCase() === 'and') {
          const nextSpace = params.nodes[index + 1];
          const secondSpace = params.nodes[index + 3];

          removeNode(nextWord);
          removeNode(nextSpace);
          removeNode(secondSpace);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***, true);

  rule.params = sortAndDedupe(getArguments(params).map(split));

  if (!rule.params.length) {
    rule.raws.afterName = '';
  ***REMOVED***
***REMOVED***

const allBugBrowers = new Set(['ie 10', 'ie 11']);

/**
 * @type {import('postcss').PluginCreator<browserslist.Options>***REMOVED***
 * @param {browserslist.Options***REMOVED*** options
 * @return {import('postcss').Plugin***REMOVED***
 */
function pluginCreator(options = {***REMOVED***) {
  const browsers = browserslist(null, {
    stats: options.stats,
    path: __dirname,
    env: options.env,
  ***REMOVED***);

  const hasAllBug = browsers.some((browser) => allBugBrowers.has(browser));
  return {
    postcssPlugin: 'postcss-minify-params',

    OnceExit(css) {
      css.walkAtRules((rule) => transform(hasAllBug, rule));
    ***REMOVED***,
  ***REMOVED***;
***REMOVED***

pluginCreator.postcss = true;
module.exports = pluginCreator;
