// postcss-normalize-repeat-style/src/index.js
'use strict';
const valueParser = require('postcss-value-parser');
const mappings = require('./lib/map');

/**
 * @param {unknown***REMOVED*** item
 * @param {number***REMOVED*** index
 * @return {boolean***REMOVED***
 */
function evenValues(item, index) {
  return index % 2 === 0;
***REMOVED***

const repeatKeywords = new Set(mappings.values());

/**
 * @param {valueParser.Node***REMOVED*** node
 * @return {boolean***REMOVED***
 */
function isCommaNode(node) {
  return node.type === 'div' && node.value === ',';
***REMOVED***

const variableFunctions = new Set(['var', 'env', 'constant']);
/**
 * @param {valueParser.Node***REMOVED*** node
 * @return {boolean***REMOVED***
 */
function isVariableFunctionNode(node) {
  if (node.type !== 'function') {
    return false;
  ***REMOVED***

  return variableFunctions.has(node.value.toLowerCase());
***REMOVED***

/**
 * @param {string***REMOVED*** value
 * @return {string***REMOVED***
 */
function transform(value) {
  const parsed = valueParser(value);

  if (parsed.nodes.length === 1) {
    return value;
  ***REMOVED***
  /** @type {{start: number?, end: number?***REMOVED***[]***REMOVED*** */
  const ranges = [];
  let rangeIndex = 0;
  let shouldContinue = true;

  parsed.nodes.forEach((node, index) => {
    // After comma (`,`) follows next background
    if (isCommaNode(node)) {
      rangeIndex += 1;
      shouldContinue = true;

      return;
    ***REMOVED***

    if (!shouldContinue) {
      return;
    ***REMOVED***

    // After separator (`/`) follows `background-size` values
    // Avoid them
    if (node.type === 'div' && node.value === '/') {
      shouldContinue = false;

      return;
    ***REMOVED***

    if (!ranges[rangeIndex]) {
      ranges[rangeIndex] = {
        start: null,
        end: null,
      ***REMOVED***;
    ***REMOVED***

    // Do not try to be processed `var and `env` function inside background
    if (isVariableFunctionNode(node)) {
      shouldContinue = false;
      ranges[rangeIndex].start = null;
      ranges[rangeIndex].end = null;

      return;
    ***REMOVED***

    const isRepeatKeyword =
      node.type === 'word' && repeatKeywords.has(node.value.toLowerCase());

    if (ranges[rangeIndex].start === null && isRepeatKeyword) {
      ranges[rangeIndex].start = index;
      ranges[rangeIndex].end = index;

      return;
    ***REMOVED***

    if (ranges[rangeIndex].start !== null) {
      if (node.type === 'space') {
        return;
      ***REMOVED*** else if (isRepeatKeyword) {
        ranges[rangeIndex].end = index;

        return;
      ***REMOVED***

      return;
    ***REMOVED***
  ***REMOVED***);

  ranges.forEach((range) => {
    if (range.start === null) {
      return;
    ***REMOVED***

    const nodes = parsed.nodes.slice(
      range.start,
      /** @type {number***REMOVED*** */ (range.end) + 1
    );

    if (nodes.length !== 3) {
      return;
    ***REMOVED***
    const key = nodes
      .filter(evenValues)
      .map((n) => n.value.toLowerCase())
      .toString();

    const match = mappings.get(key);

    if (match) {
      nodes[0].value = match;
      nodes[1].value = nodes[2].value = '';
    ***REMOVED***
  ***REMOVED***);

  return parsed.toString();
***REMOVED***

/**
 * @type {import('postcss').PluginCreator<void>***REMOVED***
 * @return {import('postcss').Plugin***REMOVED***
 */
function pluginCreator() {
  return {
    postcssPlugin: 'postcss-normalize-repeat-style',
    prepare() {
      const cache = new Map();
      return {
        OnceExit(css) {
          css.walkDecls(
            /^(background(-repeat)?|(-\w+-)?mask-repeat)$/i,
            (decl) => {
              const value = decl.value;

              if (!value) {
                return;
              ***REMOVED***

              if (cache.has(value)) {
                decl.value = cache.get(value);

                return;
              ***REMOVED***

              const result = transform(value);

              decl.value = result;
              cache.set(value, result);
            ***REMOVED***
          );
        ***REMOVED***,
      ***REMOVED***;
    ***REMOVED***,
  ***REMOVED***;
***REMOVED***

pluginCreator.postcss = true;
module.exports = pluginCreator;
