// webpack-dev-server/lib/Server.js
"use strict";

const os = require("os");
const path = require("path");
const url = require("url");
const util = require("util");
const fs = require("graceful-fs");
const ipaddr = require("ipaddr.js");
const { validate ***REMOVED*** = require("schema-utils");
const schema = require("./options.json");

/** @typedef {import("schema-utils/declarations/validate").Schema***REMOVED*** Schema */
/** @typedef {import("webpack").Compiler***REMOVED*** Compiler */
/** @typedef {import("webpack").MultiCompiler***REMOVED*** MultiCompiler */
/** @typedef {import("webpack").Configuration***REMOVED*** WebpackConfiguration */
/** @typedef {import("webpack").StatsOptions***REMOVED*** StatsOptions */
/** @typedef {import("webpack").StatsCompilation***REMOVED*** StatsCompilation */
/** @typedef {import("webpack").Stats***REMOVED*** Stats */
/** @typedef {import("webpack").MultiStats***REMOVED*** MultiStats */
/** @typedef {import("os").NetworkInterfaceInfo***REMOVED*** NetworkInterfaceInfo */
/** @typedef {import("express").Request***REMOVED*** Request */
/** @typedef {import("express").Response***REMOVED*** Response */
/** @typedef {import("express").NextFunction***REMOVED*** NextFunction */
/** @typedef {import("express").RequestHandler***REMOVED*** ExpressRequestHandler */
/** @typedef {import("express").ErrorRequestHandler***REMOVED*** ExpressErrorRequestHandler */
/** @typedef {import("chokidar").WatchOptions***REMOVED*** WatchOptions */
/** @typedef {import("chokidar").FSWatcher***REMOVED*** FSWatcher */
/** @typedef {import("connect-history-api-fallback").Options***REMOVED*** ConnectHistoryApiFallbackOptions */
/** @typedef {import("bonjour-service").Bonjour***REMOVED*** Bonjour */
/** @typedef {import("bonjour-service").Service***REMOVED*** BonjourOptions */
/** @typedef {import("http-proxy-middleware").RequestHandler***REMOVED*** RequestHandler */
/** @typedef {import("http-proxy-middleware").Options***REMOVED*** HttpProxyMiddlewareOptions */
/** @typedef {import("http-proxy-middleware").Filter***REMOVED*** HttpProxyMiddlewareOptionsFilter */
/** @typedef {import("serve-index").Options***REMOVED*** ServeIndexOptions */
/** @typedef {import("serve-static").ServeStaticOptions***REMOVED*** ServeStaticOptions */
/** @typedef {import("ipaddr.js").IPv4***REMOVED*** IPv4 */
/** @typedef {import("ipaddr.js").IPv6***REMOVED*** IPv6 */
/** @typedef {import("net").Socket***REMOVED*** Socket */
/** @typedef {import("http").IncomingMessage***REMOVED*** IncomingMessage */
/** @typedef {import("open").Options***REMOVED*** OpenOptions */

/** @typedef {import("https").ServerOptions & { spdy?: { plain?: boolean | undefined, ssl?: boolean | undefined, 'x-forwarded-for'?: string | undefined, protocol?: string | undefined, protocols?: string[] | undefined ***REMOVED******REMOVED******REMOVED*** ServerOptions */

/**
 * @template Request, Response
 * @typedef {import("webpack-dev-middleware").Options<Request, Response>***REMOVED*** DevMiddlewareOptions
 */

/**
 * @template Request, Response
 * @typedef {import("webpack-dev-middleware").Context<Request, Response>***REMOVED*** DevMiddlewareContext
 */

/**
 * @typedef {"local-ip" | "local-ipv4" | "local-ipv6" | string***REMOVED*** Host
 */

/**
 * @typedef {number | string | "auto"***REMOVED*** Port
 */

/**
 * @typedef {Object***REMOVED*** WatchFiles
 * @property {string | string[]***REMOVED*** paths
 * @property {WatchOptions & { aggregateTimeout?: number, ignored?: WatchOptions["ignored"], poll?: number | boolean ***REMOVED******REMOVED*** [options]
 */

/**
 * @typedef {Object***REMOVED*** Static
 * @property {string***REMOVED*** [directory]
 * @property {string | string[]***REMOVED*** [publicPath]
 * @property {boolean | ServeIndexOptions***REMOVED*** [serveIndex]
 * @property {ServeStaticOptions***REMOVED*** [staticOptions]
 * @property {boolean | WatchOptions & { aggregateTimeout?: number, ignored?: WatchOptions["ignored"], poll?: number | boolean ***REMOVED******REMOVED*** [watch]
 */

/**
 * @typedef {Object***REMOVED*** NormalizedStatic
 * @property {string***REMOVED*** directory
 * @property {string[]***REMOVED*** publicPath
 * @property {false | ServeIndexOptions***REMOVED*** serveIndex
 * @property {ServeStaticOptions***REMOVED*** staticOptions
 * @property {false | WatchOptions***REMOVED*** watch
 */

/**
 * @typedef {Object***REMOVED*** ServerConfiguration
 * @property {"http" | "https" | "spdy" | string***REMOVED*** [type]
 * @property {ServerOptions***REMOVED*** [options]
 */

/**
 * @typedef {Object***REMOVED*** WebSocketServerConfiguration
 * @property {"sockjs" | "ws" | string | Function***REMOVED*** [type]
 * @property {Record<string, any>***REMOVED*** [options]
 */

/**
 * @typedef {(import("ws").WebSocket | import("sockjs").Connection & { send: import("ws").WebSocket["send"], terminate: import("ws").WebSocket["terminate"], ping: import("ws").WebSocket["ping"] ***REMOVED***) & { isAlive?: boolean ***REMOVED******REMOVED*** ClientConnection
 */

/**
 * @typedef {import("ws").WebSocketServer | import("sockjs").Server & { close: import("ws").WebSocketServer["close"] ***REMOVED******REMOVED*** WebSocketServer
 */

/**
 * @typedef {{ implementation: WebSocketServer, clients: ClientConnection[] ***REMOVED******REMOVED*** WebSocketServerImplementation
 */

/**
 * @callback ByPass
 * @param {Request***REMOVED*** req
 * @param {Response***REMOVED*** res
 * @param {ProxyConfigArrayItem***REMOVED*** proxyConfig
 */

/**
 * @typedef {{ path?: HttpProxyMiddlewareOptionsFilter | undefined, context?: HttpProxyMiddlewareOptionsFilter | undefined ***REMOVED*** & { bypass?: ByPass ***REMOVED*** & HttpProxyMiddlewareOptions ***REMOVED*** ProxyConfigArrayItem
 */

/**
 * @typedef {(ProxyConfigArrayItem | ((req?: Request | undefined, res?: Response | undefined, next?: NextFunction | undefined) => ProxyConfigArrayItem))[]***REMOVED*** ProxyConfigArray
 */

/**
 * @typedef {{ [url: string]: string | ProxyConfigArrayItem ***REMOVED******REMOVED*** ProxyConfigMap
 */

/**
 * @typedef {Object***REMOVED*** OpenApp
 * @property {string***REMOVED*** [name]
 * @property {string[]***REMOVED*** [arguments]
 */

/**
 * @typedef {Object***REMOVED*** Open
 * @property {string | string[] | OpenApp***REMOVED*** [app]
 * @property {string | string[]***REMOVED*** [target]
 */

/**
 * @typedef {Object***REMOVED*** NormalizedOpen
 * @property {string***REMOVED*** target
 * @property {import("open").Options***REMOVED*** options
 */

/**
 * @typedef {Object***REMOVED*** WebSocketURL
 * @property {string***REMOVED*** [hostname]
 * @property {string***REMOVED*** [password]
 * @property {string***REMOVED*** [pathname]
 * @property {number | string***REMOVED*** [port]
 * @property {string***REMOVED*** [protocol]
 * @property {string***REMOVED*** [username]
 */

/**
 * @typedef {boolean | ((error: Error) => void)***REMOVED*** OverlayMessageOptions
 */

/**
 * @typedef {Object***REMOVED*** ClientConfiguration
 * @property {"log" | "info" | "warn" | "error" | "none" | "verbose"***REMOVED*** [logging]
 * @property {boolean  | { warnings?: OverlayMessageOptions, errors?: OverlayMessageOptions, runtimeErrors?: OverlayMessageOptions ***REMOVED******REMOVED*** [overlay]
 * @property {boolean***REMOVED*** [progress]
 * @property {boolean | number***REMOVED*** [reconnect]
 * @property {"ws" | "sockjs" | string***REMOVED*** [webSocketTransport]
 * @property {string | WebSocketURL***REMOVED*** [webSocketURL]
 */

/**
 * @typedef {Array<{ key: string; value: string ***REMOVED***> | Record<string, string | string[]>***REMOVED*** Headers
 */

/**
 * @typedef {{ name?: string, path?: string, middleware: ExpressRequestHandler | ExpressErrorRequestHandler ***REMOVED*** | ExpressRequestHandler | ExpressErrorRequestHandler***REMOVED*** Middleware
 */

/**
 * @typedef {Object***REMOVED*** Configuration
 * @property {boolean | string***REMOVED*** [ipc]
 * @property {Host***REMOVED*** [host]
 * @property {Port***REMOVED*** [port]
 * @property {boolean | "only"***REMOVED*** [hot]
 * @property {boolean***REMOVED*** [liveReload]
 * @property {DevMiddlewareOptions<Request, Response>***REMOVED*** [devMiddleware]
 * @property {boolean***REMOVED*** [compress]
 * @property {boolean***REMOVED*** [magicHtml]
 * @property {"auto" | "all" | string | string[]***REMOVED*** [allowedHosts]
 * @property {boolean | ConnectHistoryApiFallbackOptions***REMOVED*** [historyApiFallback]
 * @property {boolean | Record<string, never> | BonjourOptions***REMOVED*** [bonjour]
 * @property {string | string[] | WatchFiles | Array<string | WatchFiles>***REMOVED*** [watchFiles]
 * @property {boolean | string | Static | Array<string | Static>***REMOVED*** [static]
 * @property {boolean | ServerOptions***REMOVED*** [https]
 * @property {boolean***REMOVED*** [http2]
 * @property {"http" | "https" | "spdy" | string | ServerConfiguration***REMOVED*** [server]
 * @property {boolean | "sockjs" | "ws" | string | WebSocketServerConfiguration***REMOVED*** [webSocketServer]
 * @property {ProxyConfigMap | ProxyConfigArrayItem | ProxyConfigArray***REMOVED*** [proxy]
 * @property {boolean | string | Open | Array<string | Open>***REMOVED*** [open]
 * @property {boolean***REMOVED*** [setupExitSignals]
 * @property {boolean | ClientConfiguration***REMOVED*** [client]
 * @property {Headers | ((req: Request, res: Response, context: DevMiddlewareContext<Request, Response>) => Headers)***REMOVED*** [headers]
 * @property {(devServer: Server) => void***REMOVED*** [onAfterSetupMiddleware]
 * @property {(devServer: Server) => void***REMOVED*** [onBeforeSetupMiddleware]
 * @property {(devServer: Server) => void***REMOVED*** [onListening]
 * @property {(middlewares: Middleware[], devServer: Server) => Middleware[]***REMOVED*** [setupMiddlewares]
 */

if (!process.env.WEBPACK_SERVE) {
  // TODO fix me in the next major release
  // @ts-ignore
  process.env.WEBPACK_SERVE = true;
***REMOVED***

/**
 * @template T
 * @param fn {(function(): any) | undefined***REMOVED***
 * @returns {function(): T***REMOVED***
 */
const memoize = (fn) => {
  let cache = false;
  /** @type {T***REMOVED*** */
  let result;

  return () => {
    if (cache) {
      return result;
    ***REMOVED***

    result = /** @type {function(): any***REMOVED*** */ (fn)();
    cache = true;
    // Allow to clean up memory for fn
    // and all dependent resources
    // eslint-disable-next-line no-undefined
    fn = undefined;

    return result;
  ***REMOVED***;
***REMOVED***;

const getExpress = memoize(() => require("express"));

/**
 *
 * @param {OverlayMessageOptions***REMOVED*** [setting]
 * @returns
 */
const encodeOverlaySettings = (setting) =>
  typeof setting === "function"
    ? encodeURIComponent(setting.toString())
    : setting;

class Server {
  /**
   * @param {Configuration | Compiler | MultiCompiler***REMOVED*** options
   * @param {Compiler | MultiCompiler | Configuration***REMOVED*** compiler
   */
  constructor(options = {***REMOVED***, compiler) {
    // TODO: remove this after plugin support is published
    if (/** @type {Compiler | MultiCompiler***REMOVED*** */ (options).hooks) {
      util.deprecate(
        () => {***REMOVED***,
        "Using 'compiler' as the first argument is deprecated. Please use 'options' as the first argument and 'compiler' as the second argument.",
        "DEP_WEBPACK_DEV_SERVER_CONSTRUCTOR"
      )();

      [options = {***REMOVED***, compiler] = [compiler, options];
    ***REMOVED***

    validate(/** @type {Schema***REMOVED*** */ (schema), options, {
      name: "Dev Server",
      baseDataPath: "options",
    ***REMOVED***);

    this.compiler = /** @type {Compiler | MultiCompiler***REMOVED*** */ (compiler);
    /**
     * @type {ReturnType<Compiler["getInfrastructureLogger"]>***REMOVED***
     * */
    this.logger = this.compiler.getInfrastructureLogger("webpack-dev-server");
    this.options = /** @type {Configuration***REMOVED*** */ (options);
    /**
     * @type {FSWatcher[]***REMOVED***
     */
    this.staticWatchers = [];
    /**
     * @private
     * @type {{ name: string | symbol, listener: (...args: any[]) => void***REMOVED***[] ***REMOVED******REMOVED***
     */
    this.listeners = [];
    // Keep track of websocket proxies for external websocket upgrade.
    /**
     * @private
     * @type {RequestHandler[]***REMOVED***
     */
    this.webSocketProxies = [];
    /**
     * @type {Socket[]***REMOVED***
     */
    this.sockets = [];
    /**
     * @private
     * @type {string | undefined***REMOVED***
     */
    // eslint-disable-next-line no-undefined
    this.currentHash = undefined;
  ***REMOVED***

  // TODO compatibility with webpack v4, remove it after drop
  static get cli() {
    return {
      get getArguments() {
        return () => require("../bin/cli-flags");
      ***REMOVED***,
      get processArguments() {
        return require("../bin/process-arguments");
      ***REMOVED***,
    ***REMOVED***;
  ***REMOVED***

  static get schema() {
    return schema;
  ***REMOVED***

  /**
   * @private
   * @returns {StatsOptions***REMOVED***
   * @constructor
   */
  static get DEFAULT_STATS() {
    return {
      all: false,
      hash: true,
      warnings: true,
      errors: true,
      errorDetails: false,
    ***REMOVED***;
  ***REMOVED***

  /**
   * @param {string***REMOVED*** URL
   * @returns {boolean***REMOVED***
   */
  static isAbsoluteURL(URL) {
    // Don't match Windows paths `c:\`
    if (/^[a-zA-Z]:\\/.test(URL)) {
      return false;
    ***REMOVED***

    // Scheme: https://tools.ietf.org/html/rfc3986#section-3.1
    // Absolute URL: https://tools.ietf.org/html/rfc3986#section-4.3
    return /^[a-zA-Z][a-zA-Z\d+\-.]*:/.test(URL);
  ***REMOVED***

  /**
   * @param {string***REMOVED*** gateway
   * @returns {string | undefined***REMOVED***
   */
  static findIp(gateway) {
    const gatewayIp = ipaddr.parse(gateway);

    // Look for the matching interface in all local interfaces.
    for (const addresses of Object.values(os.networkInterfaces())) {
      for (const { cidr ***REMOVED*** of /** @type {NetworkInterfaceInfo[]***REMOVED*** */ (
        addresses
      )) {
        const net = ipaddr.parseCIDR(/** @type {string***REMOVED*** */ (cidr));

        if (
          net[0] &&
          net[0].kind() === gatewayIp.kind() &&
          gatewayIp.match(net)
        ) {
          return net[0].toString();
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  /**
   * @param {"v4" | "v6"***REMOVED*** family
   * @returns {Promise<string | undefined>***REMOVED***
   */
  static async internalIP(family) {
    try {
      const { gateway ***REMOVED*** = await require("default-gateway")[family]();
      return Server.findIp(gateway);
    ***REMOVED*** catch {
      // ignore
    ***REMOVED***
  ***REMOVED***

  /**
   * @param {"v4" | "v6"***REMOVED*** family
   * @returns {string | undefined***REMOVED***
   */
  static internalIPSync(family) {
    try {
      const { gateway ***REMOVED*** = require("default-gateway")[family].sync();
      return Server.findIp(gateway);
    ***REMOVED*** catch {
      // ignore
    ***REMOVED***
  ***REMOVED***

  /**
   * @param {Host***REMOVED*** hostname
   * @returns {Promise<string>***REMOVED***
   */
  static async getHostname(hostname) {
    if (hostname === "local-ip") {
      return (
        (await Server.internalIP("v4")) ||
        (await Server.internalIP("v6")) ||
        "0.0.0.0"
      );
    ***REMOVED*** else if (hostname === "local-ipv4") {
      return (await Server.internalIP("v4")) || "0.0.0.0";
    ***REMOVED*** else if (hostname === "local-ipv6") {
      return (await Server.internalIP("v6")) || "::";
    ***REMOVED***

    return hostname;
  ***REMOVED***

  /**
   * @param {Port***REMOVED*** port
   * @param {string***REMOVED*** host
   * @returns {Promise<number | string>***REMOVED***
   */
  static async getFreePort(port, host) {
    if (typeof port !== "undefined" && port !== null && port !== "auto") {
      return port;
    ***REMOVED***

    const pRetry = require("p-retry");
    const getPort = require("./getPort");
    const basePort =
      typeof process.env.WEBPACK_DEV_SERVER_BASE_PORT !== "undefined"
        ? parseInt(process.env.WEBPACK_DEV_SERVER_BASE_PORT, 10)
        : 8080;

    // Try to find unused port and listen on it for 3 times,
    // if port is not specified in options.
    const defaultPortRetry =
      typeof process.env.WEBPACK_DEV_SERVER_PORT_RETRY !== "undefined"
        ? parseInt(process.env.WEBPACK_DEV_SERVER_PORT_RETRY, 10)
        : 3;

    return pRetry(() => getPort(basePort, host), {
      retries: defaultPortRetry,
    ***REMOVED***);
  ***REMOVED***

  /**
   * @returns {string***REMOVED***
   */
  static findCacheDir() {
    const cwd = process.cwd();

    /**
     * @type {string | undefined***REMOVED***
     */
    let dir = cwd;

    for (;;) {
      try {
        if (fs.statSync(path.join(dir, "package.json")).isFile()) break;
        // eslint-disable-next-line no-empty
      ***REMOVED*** catch (e) {***REMOVED***

      const parent = path.dirname(dir);

      if (dir === parent) {
        // eslint-disable-next-line no-undefined
        dir = undefined;
        break;
      ***REMOVED***

      dir = parent;
    ***REMOVED***

    if (!dir) {
      return path.resolve(cwd, ".cache/webpack-dev-server");
    ***REMOVED*** else if (process.versions.pnp === "1") {
      return path.resolve(dir, ".pnp/.cache/webpack-dev-server");
    ***REMOVED*** else if (process.versions.pnp === "3") {
      return path.resolve(dir, ".yarn/.cache/webpack-dev-server");
    ***REMOVED***

    return path.resolve(dir, "node_modules/.cache/webpack-dev-server");
  ***REMOVED***

  /**
   * @private
   * @param {Compiler***REMOVED*** compiler
   * @returns bool
   */
  static isWebTarget(compiler) {
    // TODO improve for the next major version - we should store `web` and other targets in `compiler.options.environment`
    if (
      compiler.options.externalsPresets &&
      compiler.options.externalsPresets.web
    ) {
      return true;
    ***REMOVED***

    if (
      compiler.options.resolve.conditionNames &&
      compiler.options.resolve.conditionNames.includes("browser")
    ) {
      return true;
    ***REMOVED***

    const webTargets = [
      "web",
      "webworker",
      "electron-preload",
      "electron-renderer",
      "node-webkit",
      // eslint-disable-next-line no-undefined
      undefined,
      null,
    ];

    if (Array.isArray(compiler.options.target)) {
      return compiler.options.target.some((r) => webTargets.includes(r));
    ***REMOVED***

    return webTargets.includes(/** @type {string***REMOVED*** */ (compiler.options.target));
  ***REMOVED***

  /**
   * @private
   * @param {Compiler***REMOVED*** compiler
   */
  addAdditionalEntries(compiler) {
    /**
     * @type {string[]***REMOVED***
     */
    const additionalEntries = [];
    const isWebTarget = Server.isWebTarget(compiler);

    // TODO maybe empty client
    if (this.options.client && isWebTarget) {
      let webSocketURLStr = "";

      if (this.options.webSocketServer) {
        const webSocketURL =
          /** @type {WebSocketURL***REMOVED*** */
          (
            /** @type {ClientConfiguration***REMOVED*** */
            (this.options.client).webSocketURL
          );
        const webSocketServer =
          /** @type {{ type: WebSocketServerConfiguration["type"], options: NonNullable<WebSocketServerConfiguration["options"]> ***REMOVED******REMOVED*** */
          (this.options.webSocketServer);
        const searchParams = new URLSearchParams();

        /** @type {string***REMOVED*** */
        let protocol;

        // We are proxying dev server and need to specify custom `hostname`
        if (typeof webSocketURL.protocol !== "undefined") {
          protocol = webSocketURL.protocol;
        ***REMOVED*** else {
          protocol =
            /** @type {ServerConfiguration***REMOVED*** */
            (this.options.server).type === "http" ? "ws:" : "wss:";
        ***REMOVED***

        searchParams.set("protocol", protocol);

        if (typeof webSocketURL.username !== "undefined") {
          searchParams.set("username", webSocketURL.username);
        ***REMOVED***

        if (typeof webSocketURL.password !== "undefined") {
          searchParams.set("password", webSocketURL.password);
        ***REMOVED***

        /** @type {string***REMOVED*** */
        let hostname;

        // SockJS is not supported server mode, so `hostname` and `port` can't specified, let's ignore them
        // TODO show warning about this
        const isSockJSType = webSocketServer.type === "sockjs";

        // We are proxying dev server and need to specify custom `hostname`
        if (typeof webSocketURL.hostname !== "undefined") {
          hostname = webSocketURL.hostname;
        ***REMOVED***
        // Web socket server works on custom `hostname`, only for `ws` because `sock-js` is not support custom `hostname`
        else if (
          typeof webSocketServer.options.host !== "undefined" &&
          !isSockJSType
        ) {
          hostname = webSocketServer.options.host;
        ***REMOVED***
        // The `host` option is specified
        else if (typeof this.options.host !== "undefined") {
          hostname = this.options.host;
        ***REMOVED***
        // The `port` option is not specified
        else {
          hostname = "0.0.0.0";
        ***REMOVED***

        searchParams.set("hostname", hostname);

        /** @type {number | string***REMOVED*** */
        let port;

        // We are proxying dev server and need to specify custom `port`
        if (typeof webSocketURL.port !== "undefined") {
          port = webSocketURL.port;
        ***REMOVED***
        // Web socket server works on custom `port`, only for `ws` because `sock-js` is not support custom `port`
        else if (
          typeof webSocketServer.options.port !== "undefined" &&
          !isSockJSType
        ) {
          port = webSocketServer.options.port;
        ***REMOVED***
        // The `port` option is specified
        else if (typeof this.options.port === "number") {
          port = this.options.port;
        ***REMOVED***
        // The `port` option is specified using `string`
        else if (
          typeof this.options.port === "string" &&
          this.options.port !== "auto"
        ) {
          port = Number(this.options.port);
        ***REMOVED***
        // The `port` option is not specified or set to `auto`
        else {
          port = "0";
        ***REMOVED***

        searchParams.set("port", String(port));

        /** @type {string***REMOVED*** */
        let pathname = "";

        // We are proxying dev server and need to specify custom `pathname`
        if (typeof webSocketURL.pathname !== "undefined") {
          pathname = webSocketURL.pathname;
        ***REMOVED***
        // Web socket server works on custom `path`
        else if (
          typeof webSocketServer.options.prefix !== "undefined" ||
          typeof webSocketServer.options.path !== "undefined"
        ) {
          pathname =
            webSocketServer.options.prefix || webSocketServer.options.path;
        ***REMOVED***

        searchParams.set("pathname", pathname);

        const client = /** @type {ClientConfiguration***REMOVED*** */ (this.options.client);

        if (typeof client.logging !== "undefined") {
          searchParams.set("logging", client.logging);
        ***REMOVED***

        if (typeof client.progress !== "undefined") {
          searchParams.set("progress", String(client.progress));
        ***REMOVED***

        if (typeof client.overlay !== "undefined") {
          const overlayString =
            typeof client.overlay === "boolean"
              ? String(client.overlay)
              : JSON.stringify({
                  ...client.overlay,
                  errors: encodeOverlaySettings(client.overlay.errors),
                  warnings: encodeOverlaySettings(client.overlay.warnings),
                  runtimeErrors: encodeOverlaySettings(
                    client.overlay.runtimeErrors
                  ),
                ***REMOVED***);

          searchParams.set("overlay", overlayString);
        ***REMOVED***

        if (typeof client.reconnect !== "undefined") {
          searchParams.set(
            "reconnect",
            typeof client.reconnect === "number"
              ? String(client.reconnect)
              : "10"
          );
        ***REMOVED***

        if (typeof this.options.hot !== "undefined") {
          searchParams.set("hot", String(this.options.hot));
        ***REMOVED***

        if (typeof this.options.liveReload !== "undefined") {
          searchParams.set("live-reload", String(this.options.liveReload));
        ***REMOVED***

        webSocketURLStr = searchParams.toString();
      ***REMOVED***

      additionalEntries.push(
        `${require.resolve("../client/index.js")***REMOVED***?${webSocketURLStr***REMOVED***`
      );
    ***REMOVED***

    if (this.options.hot === "only") {
      additionalEntries.push(require.resolve("webpack/hot/only-dev-server"));
    ***REMOVED*** else if (this.options.hot) {
      additionalEntries.push(require.resolve("webpack/hot/dev-server"));
    ***REMOVED***

    const webpack = compiler.webpack || require("webpack");

    // use a hook to add entries if available
    if (typeof webpack.EntryPlugin !== "undefined") {
      for (const additionalEntry of additionalEntries) {
        new webpack.EntryPlugin(compiler.context, additionalEntry, {
          // eslint-disable-next-line no-undefined
          name: undefined,
        ***REMOVED***).apply(compiler);
      ***REMOVED***
    ***REMOVED***
    // TODO remove after drop webpack v4 support
    else {
      /**
       * prependEntry Method for webpack 4
       * @param {any***REMOVED*** originalEntry
       * @param {any***REMOVED*** newAdditionalEntries
       * @returns {any***REMOVED***
       */
      const prependEntry = (originalEntry, newAdditionalEntries) => {
        if (typeof originalEntry === "function") {
          return () =>
            Promise.resolve(originalEntry()).then((entry) =>
              prependEntry(entry, newAdditionalEntries)
            );
        ***REMOVED***

        if (
          typeof originalEntry === "object" &&
          !Array.isArray(originalEntry)
        ) {
          /** @type {Object<string,string>***REMOVED*** */
          const clone = {***REMOVED***;

          Object.keys(originalEntry).forEach((key) => {
            // entry[key] should be a string here
            const entryDescription = originalEntry[key];

            clone[key] = prependEntry(entryDescription, newAdditionalEntries);
          ***REMOVED***);

          return clone;
        ***REMOVED***

        // in this case, entry is a string or an array.
        // make sure that we do not add duplicates.
        /** @type {any***REMOVED*** */
        const entriesClone = additionalEntries.slice(0);

        [].concat(originalEntry).forEach((newEntry) => {
          if (!entriesClone.includes(newEntry)) {
            entriesClone.push(newEntry);
          ***REMOVED***
        ***REMOVED***);

        return entriesClone;
      ***REMOVED***;

      compiler.options.entry = prependEntry(
        compiler.options.entry || "./src",
        additionalEntries
      );
      compiler.hooks.entryOption.call(
        /** @type {string***REMOVED*** */ (compiler.options.context),
        compiler.options.entry
      );
    ***REMOVED***
  ***REMOVED***

  /**
   * @private
   * @returns {Compiler["options"]***REMOVED***
   */
  getCompilerOptions() {
    if (
      typeof (/** @type {MultiCompiler***REMOVED*** */ (this.compiler).compilers) !==
      "undefined"
    ) {
      if (/** @type {MultiCompiler***REMOVED*** */ (this.compiler).compilers.length === 1) {
        return (
          /** @type {MultiCompiler***REMOVED*** */
          (this.compiler).compilers[0].options
        );
      ***REMOVED***

      // Configuration with the `devServer` options
      const compilerWithDevServer =
        /** @type {MultiCompiler***REMOVED*** */
        (this.compiler).compilers.find((config) => config.options.devServer);

      if (compilerWithDevServer) {
        return compilerWithDevServer.options;
      ***REMOVED***

      // Configuration with `web` preset
      const compilerWithWebPreset =
        /** @type {MultiCompiler***REMOVED*** */
        (this.compiler).compilers.find(
          (config) =>
            (config.options.externalsPresets &&
              config.options.externalsPresets.web) ||
            [
              "web",
              "webworker",
              "electron-preload",
              "electron-renderer",
              "node-webkit",
              // eslint-disable-next-line no-undefined
              undefined,
              null,
            ].includes(/** @type {string***REMOVED*** */ (config.options.target))
        );

      if (compilerWithWebPreset) {
        return compilerWithWebPreset.options;
      ***REMOVED***

      // Fallback
      return /** @type {MultiCompiler***REMOVED*** */ (this.compiler).compilers[0].options;
    ***REMOVED***

    return /** @type {Compiler***REMOVED*** */ (this.compiler).options;
  ***REMOVED***

  /**
   * @private
   * @returns {Promise<void>***REMOVED***
   */
  async normalizeOptions() {
    const { options ***REMOVED*** = this;
    const compilerOptions = this.getCompilerOptions();
    // TODO remove `{***REMOVED***` after drop webpack v4 support
    const compilerWatchOptions = compilerOptions.watchOptions || {***REMOVED***;
    /**
     * @param {WatchOptions & { aggregateTimeout?: number, ignored?: WatchOptions["ignored"], poll?: number | boolean ***REMOVED******REMOVED*** watchOptions
     * @returns {WatchOptions***REMOVED***
     */
    const getWatchOptions = (watchOptions = {***REMOVED***) => {
      const getPolling = () => {
        if (typeof watchOptions.usePolling !== "undefined") {
          return watchOptions.usePolling;
        ***REMOVED***

        if (typeof watchOptions.poll !== "undefined") {
          return Boolean(watchOptions.poll);
        ***REMOVED***

        if (typeof compilerWatchOptions.poll !== "undefined") {
          return Boolean(compilerWatchOptions.poll);
        ***REMOVED***

        return false;
      ***REMOVED***;
      const getInterval = () => {
        if (typeof watchOptions.interval !== "undefined") {
          return watchOptions.interval;
        ***REMOVED***

        if (typeof watchOptions.poll === "number") {
          return watchOptions.poll;
        ***REMOVED***

        if (typeof compilerWatchOptions.poll === "number") {
          return compilerWatchOptions.poll;
        ***REMOVED***
      ***REMOVED***;

      const usePolling = getPolling();
      const interval = getInterval();
      const { poll, ...rest ***REMOVED*** = watchOptions;

      return {
        ignoreInitial: true,
        persistent: true,
        followSymlinks: false,
        atomic: false,
        alwaysStat: true,
        ignorePermissionErrors: true,
        // Respect options from compiler watchOptions
        usePolling,
        interval,
        ignored: watchOptions.ignored,
        // TODO: we respect these options for all watch options and allow developers to pass them to chokidar, but chokidar doesn't have these options maybe we need revisit that in future
        ...rest,
      ***REMOVED***;
    ***REMOVED***;
    /**
     * @param {string | Static | undefined***REMOVED*** [optionsForStatic]
     * @returns {NormalizedStatic***REMOVED***
     */
    const getStaticItem = (optionsForStatic) => {
      const getDefaultStaticOptions = () => {
        return {
          directory: path.join(process.cwd(), "public"),
          staticOptions: {***REMOVED***,
          publicPath: ["/"],
          serveIndex: { icons: true ***REMOVED***,
          watch: getWatchOptions(),
        ***REMOVED***;
      ***REMOVED***;

      /** @type {NormalizedStatic***REMOVED*** */
      let item;

      if (typeof optionsForStatic === "undefined") {
        item = getDefaultStaticOptions();
      ***REMOVED*** else if (typeof optionsForStatic === "string") {
        item = {
          ...getDefaultStaticOptions(),
          directory: optionsForStatic,
        ***REMOVED***;
      ***REMOVED*** else {
        const def = getDefaultStaticOptions();

        item = {
          directory:
            typeof optionsForStatic.directory !== "undefined"
              ? optionsForStatic.directory
              : def.directory,
          // TODO: do merge in the next major release
          staticOptions:
            typeof optionsForStatic.staticOptions !== "undefined"
              ? optionsForStatic.staticOptions
              : def.staticOptions,
          publicPath:
            // eslint-disable-next-line no-nested-ternary
            typeof optionsForStatic.publicPath !== "undefined"
              ? Array.isArray(optionsForStatic.publicPath)
                ? optionsForStatic.publicPath
                : [optionsForStatic.publicPath]
              : def.publicPath,
          // TODO: do merge in the next major release
          serveIndex:
            // eslint-disable-next-line no-nested-ternary
            typeof optionsForStatic.serveIndex !== "undefined"
              ? typeof optionsForStatic.serveIndex === "boolean" &&
                optionsForStatic.serveIndex
                ? def.serveIndex
                : optionsForStatic.serveIndex
              : def.serveIndex,
          watch:
            // eslint-disable-next-line no-nested-ternary
            typeof optionsForStatic.watch !== "undefined"
              ? // eslint-disable-next-line no-nested-ternary
                typeof optionsForStatic.watch === "boolean"
                ? optionsForStatic.watch
                  ? def.watch
                  : false
                : getWatchOptions(optionsForStatic.watch)
              : def.watch,
        ***REMOVED***;
      ***REMOVED***

      if (Server.isAbsoluteURL(item.directory)) {
        throw new Error("Using a URL as static.directory is not supported");
      ***REMOVED***

      return item;
    ***REMOVED***;

    if (typeof options.allowedHosts === "undefined") {
      // AllowedHosts allows some default hosts picked from `options.host` or `webSocketURL.hostname` and `localhost`
      options.allowedHosts = "auto";
    ***REMOVED***
    // We store allowedHosts as array when supplied as string
    else if (
      typeof options.allowedHosts === "string" &&
      options.allowedHosts !== "auto" &&
      options.allowedHosts !== "all"
    ) {
      options.allowedHosts = [options.allowedHosts];
    ***REMOVED***
    // CLI pass options as array, we should normalize them
    else if (
      Array.isArray(options.allowedHosts) &&
      options.allowedHosts.includes("all")
    ) {
      options.allowedHosts = "all";
    ***REMOVED***

    if (typeof options.bonjour === "undefined") {
      options.bonjour = false;
    ***REMOVED*** else if (typeof options.bonjour === "boolean") {
      options.bonjour = options.bonjour ? {***REMOVED*** : false;
    ***REMOVED***

    if (
      typeof options.client === "undefined" ||
      (typeof options.client === "object" && options.client !== null)
    ) {
      if (!options.client) {
        options.client = {***REMOVED***;
      ***REMOVED***

      if (typeof options.client.webSocketURL === "undefined") {
        options.client.webSocketURL = {***REMOVED***;
      ***REMOVED*** else if (typeof options.client.webSocketURL === "string") {
        const parsedURL = new URL(options.client.webSocketURL);

        options.client.webSocketURL = {
          protocol: parsedURL.protocol,
          hostname: parsedURL.hostname,
          port: parsedURL.port.length > 0 ? Number(parsedURL.port) : "",
          pathname: parsedURL.pathname,
          username: parsedURL.username,
          password: parsedURL.password,
        ***REMOVED***;
      ***REMOVED*** else if (typeof options.client.webSocketURL.port === "string") {
        options.client.webSocketURL.port = Number(
          options.client.webSocketURL.port
        );
      ***REMOVED***

      // Enable client overlay by default
      if (typeof options.client.overlay === "undefined") {
        options.client.overlay = true;
      ***REMOVED*** else if (typeof options.client.overlay !== "boolean") {
        options.client.overlay = {
          errors: true,
          warnings: true,
          ...options.client.overlay,
        ***REMOVED***;
      ***REMOVED***

      if (typeof options.client.reconnect === "undefined") {
        options.client.reconnect = 10;
      ***REMOVED*** else if (options.client.reconnect === true) {
        options.client.reconnect = Infinity;
      ***REMOVED*** else if (options.client.reconnect === false) {
        options.client.reconnect = 0;
      ***REMOVED***

      // Respect infrastructureLogging.level
      if (typeof options.client.logging === "undefined") {
        options.client.logging = compilerOptions.infrastructureLogging
          ? compilerOptions.infrastructureLogging.level
          : "info";
      ***REMOVED***
    ***REMOVED***

    if (typeof options.compress === "undefined") {
      options.compress = true;
    ***REMOVED***

    if (typeof options.devMiddleware === "undefined") {
      options.devMiddleware = {***REMOVED***;
    ***REMOVED***

    // No need to normalize `headers`

    if (typeof options.historyApiFallback === "undefined") {
      options.historyApiFallback = false;
    ***REMOVED*** else if (
      typeof options.historyApiFallback === "boolean" &&
      options.historyApiFallback
    ) {
      options.historyApiFallback = {***REMOVED***;
    ***REMOVED***

    // No need to normalize `host`

    options.hot =
      typeof options.hot === "boolean" || options.hot === "only"
        ? options.hot
        : true;

    const isHTTPs = Boolean(options.https);
    const isSPDY = Boolean(options.http2);

    if (isHTTPs) {
      // TODO: remove in the next major release
      util.deprecate(
        () => {***REMOVED***,
        "'https' option is deprecated. Please use the 'server' option.",
        "DEP_WEBPACK_DEV_SERVER_HTTPS"
      )();
    ***REMOVED***

    if (isSPDY) {
      // TODO: remove in the next major release
      util.deprecate(
        () => {***REMOVED***,
        "'http2' option is deprecated. Please use the 'server' option.",
        "DEP_WEBPACK_DEV_SERVER_HTTP2"
      )();
    ***REMOVED***

    options.server = {
      type:
        // eslint-disable-next-line no-nested-ternary
        typeof options.server === "string"
          ? options.server
          : // eslint-disable-next-line no-nested-ternary
          typeof (options.server || {***REMOVED***).type === "string"
          ? /** @type {ServerConfiguration***REMOVED*** */ (options.server).type || "http"
          : // eslint-disable-next-line no-nested-ternary
          isSPDY
          ? "spdy"
          : isHTTPs
          ? "https"
          : "http",
      options: {
        .../** @type {ServerOptions***REMOVED*** */ (options.https),
        .../** @type {ServerConfiguration***REMOVED*** */ (options.server || {***REMOVED***).options,
      ***REMOVED***,
    ***REMOVED***;

    if (
      options.server.type === "spdy" &&
      typeof (/** @type {ServerOptions***REMOVED*** */ (options.server.options).spdy) ===
        "undefined"
    ) {
      /** @type {ServerOptions***REMOVED*** */
      (options.server.options).spdy = {
        protocols: ["h2", "http/1.1"],
      ***REMOVED***;
    ***REMOVED***

    if (options.server.type === "https" || options.server.type === "spdy") {
      if (
        typeof (
          /** @type {ServerOptions***REMOVED*** */ (options.server.options).requestCert
        ) === "undefined"
      ) {
        /** @type {ServerOptions***REMOVED*** */
        (options.server.options).requestCert = false;
      ***REMOVED***

      const httpsProperties =
        /** @type {Array<keyof ServerOptions>***REMOVED*** */
        (["cacert", "ca", "cert", "crl", "key", "pfx"]);

      for (const property of httpsProperties) {
        if (
          typeof (
            /** @type {ServerOptions***REMOVED*** */ (options.server.options)[property]
          ) === "undefined"
        ) {
          // eslint-disable-next-line no-continue
          continue;
        ***REMOVED***

        // @ts-ignore
        if (property === "cacert") {
          // TODO remove the `cacert` option in favor `ca` in the next major release
          util.deprecate(
            () => {***REMOVED***,
            "The 'cacert' option is deprecated. Please use the 'ca' option.",
            "DEP_WEBPACK_DEV_SERVER_CACERT"
          )();
        ***REMOVED***

        /** @type {any***REMOVED*** */
        const value =
          /** @type {ServerOptions***REMOVED*** */
          (options.server.options)[property];
        /**
         * @param {string | Buffer | undefined***REMOVED*** item
         * @returns {string | Buffer | undefined***REMOVED***
         */
        const readFile = (item) => {
          if (
            Buffer.isBuffer(item) ||
            (typeof item === "object" && item !== null && !Array.isArray(item))
          ) {
            return item;
          ***REMOVED***

          if (item) {
            let stats = null;

            try {
              stats = fs.lstatSync(fs.realpathSync(item)).isFile();
            ***REMOVED*** catch (error) {
              // Ignore error
            ***REMOVED***

            // It is a file
            return stats ? fs.readFileSync(item) : item;
          ***REMOVED***
        ***REMOVED***;

        /** @type {any***REMOVED*** */
        (options.server.options)[property] = Array.isArray(value)
          ? value.map((item) => readFile(item))
          : readFile(value);
      ***REMOVED***

      let fakeCert;

      if (
        !(/** @type {ServerOptions***REMOVED*** */ (options.server.options).key) ||
        !(/** @type {ServerOptions***REMOVED*** */ (options.server.options).cert)
      ) {
        const certificateDir = Server.findCacheDir();
        const certificatePath = path.join(certificateDir, "server.pem");
        let certificateExists;

        try {
          const certificate = await fs.promises.stat(certificatePath);
          certificateExists = certificate.isFile();
        ***REMOVED*** catch {
          certificateExists = false;
        ***REMOVED***

        if (certificateExists) {
          const certificateTtl = 1000 * 60 * 60 * 24;
          const certificateStat = await fs.promises.stat(certificatePath);
          const now = Number(new Date());

          // cert is more than 30 days old, kill it with fire
          if ((now - Number(certificateStat.ctime)) / certificateTtl > 30) {
            const { promisify ***REMOVED*** = require("util");
            const rimraf = require("rimraf");
            const del = promisify(rimraf);

            this.logger.info(
              "SSL certificate is more than 30 days old. Removing..."
            );

            await del(certificatePath);

            certificateExists = false;
          ***REMOVED***
        ***REMOVED***

        if (!certificateExists) {
          this.logger.info("Generating SSL certificate...");

          // @ts-ignore
          const selfsigned = require("selfsigned");
          const attributes = [{ name: "commonName", value: "localhost" ***REMOVED***];
          const pems = selfsigned.generate(attributes, {
            algorithm: "sha256",
            days: 30,
            keySize: 2048,
            extensions: [
              {
                name: "basicConstraints",
                cA: true,
              ***REMOVED***,
              {
                name: "keyUsage",
                keyCertSign: true,
                digitalSignature: true,
                nonRepudiation: true,
                keyEncipherment: true,
                dataEncipherment: true,
              ***REMOVED***,
              {
                name: "extKeyUsage",
                serverAuth: true,
                clientAuth: true,
                codeSigning: true,
                timeStamping: true,
              ***REMOVED***,
              {
                name: "subjectAltName",
                altNames: [
                  {
                    // type 2 is DNS
                    type: 2,
                    value: "localhost",
                  ***REMOVED***,
                  {
                    type: 2,
                    value: "localhost.localdomain",
                  ***REMOVED***,
                  {
                    type: 2,
                    value: "lvh.me",
                  ***REMOVED***,
                  {
                    type: 2,
                    value: "*.lvh.me",
                  ***REMOVED***,
                  {
                    type: 2,
                    value: "[::1]",
                  ***REMOVED***,
                  {
                    // type 7 is IP
                    type: 7,
                    ip: "127.0.0.1",
                  ***REMOVED***,
                  {
                    type: 7,
                    ip: "fe80::1",
                  ***REMOVED***,
                ],
              ***REMOVED***,
            ],
          ***REMOVED***);

          await fs.promises.mkdir(certificateDir, { recursive: true ***REMOVED***);

          await fs.promises.writeFile(
            certificatePath,
            pems.private + pems.cert,
            {
              encoding: "utf8",
            ***REMOVED***
          );
        ***REMOVED***

        fakeCert = await fs.promises.readFile(certificatePath);

        this.logger.info(`SSL certificate: ${certificatePath***REMOVED***`);
      ***REMOVED***

      if (
        /** @type {ServerOptions & { cacert?: ServerOptions["ca"] ***REMOVED******REMOVED*** */ (
          options.server.options
        ).cacert
      ) {
        if (/** @type {ServerOptions***REMOVED*** */ (options.server.options).ca) {
          this.logger.warn(
            "Do not specify 'ca' and 'cacert' options together, the 'ca' option will be used."
          );
        ***REMOVED*** else {
          /** @type {ServerOptions***REMOVED*** */
          (options.server.options).ca =
            /** @type {ServerOptions & { cacert?: ServerOptions["ca"] ***REMOVED******REMOVED*** */
            (options.server.options).cacert;
        ***REMOVED***

        delete (
          /** @type {ServerOptions & { cacert?: ServerOptions["ca"] ***REMOVED******REMOVED*** */ (
            options.server.options
          ).cacert
        );
      ***REMOVED***

      /** @type {ServerOptions***REMOVED*** */
      (options.server.options).key =
        /** @type {ServerOptions***REMOVED*** */
        (options.server.options).key || fakeCert;
      /** @type {ServerOptions***REMOVED*** */
      (options.server.options).cert =
        /** @type {ServerOptions***REMOVED*** */
        (options.server.options).cert || fakeCert;
    ***REMOVED***

    if (typeof options.ipc === "boolean") {
      const isWindows = process.platform === "win32";
      const pipePrefix = isWindows ? "\\\\.\\pipe\\" : os.tmpdir();
      const pipeName = "webpack-dev-server.sock";

      options.ipc = path.join(pipePrefix, pipeName);
    ***REMOVED***

    options.liveReload =
      typeof options.liveReload !== "undefined" ? options.liveReload : true;

    options.magicHtml =
      typeof options.magicHtml !== "undefined" ? options.magicHtml : true;

    // https://github.com/webpack/webpack-dev-server/issues/1990
    const defaultOpenOptions = { wait: false ***REMOVED***;
    /**
     * @param {any***REMOVED*** target
     * @returns {NormalizedOpen[]***REMOVED***
     */
    // TODO: remove --open-app in favor of --open-app-name
    const getOpenItemsFromObject = ({ target, ...rest ***REMOVED***) => {
      const normalizedOptions = { ...defaultOpenOptions, ...rest ***REMOVED***;

      if (typeof normalizedOptions.app === "string") {
        normalizedOptions.app = {
          name: normalizedOptions.app,
        ***REMOVED***;
      ***REMOVED***

      const normalizedTarget = typeof target === "undefined" ? "<url>" : target;

      if (Array.isArray(normalizedTarget)) {
        return normalizedTarget.map((singleTarget) => {
          return { target: singleTarget, options: normalizedOptions ***REMOVED***;
        ***REMOVED***);
      ***REMOVED***

      return [{ target: normalizedTarget, options: normalizedOptions ***REMOVED***];
    ***REMOVED***;

    if (typeof options.open === "undefined") {
      /** @type {NormalizedOpen[]***REMOVED*** */
      (options.open) = [];
    ***REMOVED*** else if (typeof options.open === "boolean") {
      /** @type {NormalizedOpen[]***REMOVED*** */
      (options.open) = options.open
        ? [
            {
              target: "<url>",
              options: /** @type {OpenOptions***REMOVED*** */ (defaultOpenOptions),
            ***REMOVED***,
          ]
        : [];
    ***REMOVED*** else if (typeof options.open === "string") {
      /** @type {NormalizedOpen[]***REMOVED*** */
      (options.open) = [{ target: options.open, options: defaultOpenOptions ***REMOVED***];
    ***REMOVED*** else if (Array.isArray(options.open)) {
      /**
       * @type {NormalizedOpen[]***REMOVED***
       */
      const result = [];

      options.open.forEach((item) => {
        if (typeof item === "string") {
          result.push({ target: item, options: defaultOpenOptions ***REMOVED***);

          return;
        ***REMOVED***

        result.push(...getOpenItemsFromObject(item));
      ***REMOVED***);

      /** @type {NormalizedOpen[]***REMOVED*** */
      (options.open) = result;
    ***REMOVED*** else {
      /** @type {NormalizedOpen[]***REMOVED*** */
      (options.open) = [...getOpenItemsFromObject(options.open)];
    ***REMOVED***

    if (options.onAfterSetupMiddleware) {
      // TODO: remove in the next major release
      util.deprecate(
        () => {***REMOVED***,
        "'onAfterSetupMiddleware' option is deprecated. Please use the 'setupMiddlewares' option.",
        `DEP_WEBPACK_DEV_SERVER_ON_AFTER_SETUP_MIDDLEWARE`
      )();
    ***REMOVED***

    if (options.onBeforeSetupMiddleware) {
      // TODO: remove in the next major release
      util.deprecate(
        () => {***REMOVED***,
        "'onBeforeSetupMiddleware' option is deprecated. Please use the 'setupMiddlewares' option.",
        `DEP_WEBPACK_DEV_SERVER_ON_BEFORE_SETUP_MIDDLEWARE`
      )();
    ***REMOVED***

    if (typeof options.port === "string" && options.port !== "auto") {
      options.port = Number(options.port);
    ***REMOVED***

    /**
     * Assume a proxy configuration specified as:
     * proxy: {
     *   'context': { options ***REMOVED***
     * ***REMOVED***
     * OR
     * proxy: {
     *   'context': 'target'
     * ***REMOVED***
     */
    if (typeof options.proxy !== "undefined") {
      // TODO remove in the next major release, only accept `Array`
      if (!Array.isArray(options.proxy)) {
        if (
          Object.prototype.hasOwnProperty.call(options.proxy, "target") ||
          Object.prototype.hasOwnProperty.call(options.proxy, "router")
        ) {
          /** @type {ProxyConfigArray***REMOVED*** */
          (options.proxy) = [/** @type {ProxyConfigMap***REMOVED*** */ (options.proxy)];
        ***REMOVED*** else {
          /** @type {ProxyConfigArray***REMOVED*** */
          (options.proxy) = Object.keys(options.proxy).map(
            /**
             * @param {string***REMOVED*** context
             * @returns {HttpProxyMiddlewareOptions***REMOVED***
             */
            (context) => {
              let proxyOptions;
              // For backwards compatibility reasons.
              const correctedContext = context
                .replace(/^\*$/, "**")
                .replace(/\/\*$/, "");

              if (
                typeof (
                  /** @type {ProxyConfigMap***REMOVED*** */ (options.proxy)[context]
                ) === "string"
              ) {
                proxyOptions = {
                  context: correctedContext,
                  target:
                    /** @type {ProxyConfigMap***REMOVED*** */
                    (options.proxy)[context],
                ***REMOVED***;
              ***REMOVED*** else {
                proxyOptions = {
                  // @ts-ignore
                  .../** @type {ProxyConfigMap***REMOVED*** */ (options.proxy)[context],
                ***REMOVED***;
                proxyOptions.context = correctedContext;
              ***REMOVED***

              return proxyOptions;
            ***REMOVED***
          );
        ***REMOVED***
      ***REMOVED***

      /** @type {ProxyConfigArray***REMOVED*** */
      (options.proxy) =
        /** @type {ProxyConfigArray***REMOVED*** */
        (options.proxy).map((item) => {
          if (typeof item === "function") {
            return item;
          ***REMOVED***

          /**
           * @param {"info" | "warn" | "error" | "debug" | "silent" | undefined | "none" | "log" | "verbose"***REMOVED*** level
           * @returns {"info" | "warn" | "error" | "debug" | "silent" | undefined***REMOVED***
           */
          const getLogLevelForProxy = (level) => {
            if (level === "none") {
              return "silent";
            ***REMOVED***

            if (level === "log") {
              return "info";
            ***REMOVED***

            if (level === "verbose") {
              return "debug";
            ***REMOVED***

            return level;
          ***REMOVED***;

          if (typeof item.logLevel === "undefined") {
            item.logLevel = getLogLevelForProxy(
              compilerOptions.infrastructureLogging
                ? compilerOptions.infrastructureLogging.level
                : "info"
            );
          ***REMOVED***

          if (typeof item.logProvider === "undefined") {
            item.logProvider = () => this.logger;
          ***REMOVED***

          return item;
        ***REMOVED***);
    ***REMOVED***

    if (typeof options.setupExitSignals === "undefined") {
      options.setupExitSignals = true;
    ***REMOVED***

    if (typeof options.static === "undefined") {
      options.static = [getStaticItem()];
    ***REMOVED*** else if (typeof options.static === "boolean") {
      options.static = options.static ? [getStaticItem()] : false;
    ***REMOVED*** else if (typeof options.static === "string") {
      options.static = [getStaticItem(options.static)];
    ***REMOVED*** else if (Array.isArray(options.static)) {
      options.static = options.static.map((item) => getStaticItem(item));
    ***REMOVED*** else {
      options.static = [getStaticItem(options.static)];
    ***REMOVED***

    if (typeof options.watchFiles === "string") {
      options.watchFiles = [
        { paths: options.watchFiles, options: getWatchOptions() ***REMOVED***,
      ];
    ***REMOVED*** else if (
      typeof options.watchFiles === "object" &&
      options.watchFiles !== null &&
      !Array.isArray(options.watchFiles)
    ) {
      options.watchFiles = [
        {
          paths: options.watchFiles.paths,
          options: getWatchOptions(options.watchFiles.options || {***REMOVED***),
        ***REMOVED***,
      ];
    ***REMOVED*** else if (Array.isArray(options.watchFiles)) {
      options.watchFiles = options.watchFiles.map((item) => {
        if (typeof item === "string") {
          return { paths: item, options: getWatchOptions() ***REMOVED***;
        ***REMOVED***

        return {
          paths: item.paths,
          options: getWatchOptions(item.options || {***REMOVED***),
        ***REMOVED***;
      ***REMOVED***);
    ***REMOVED*** else {
      options.watchFiles = [];
    ***REMOVED***

    const defaultWebSocketServerType = "ws";
    const defaultWebSocketServerOptions = { path: "/ws" ***REMOVED***;

    if (typeof options.webSocketServer === "undefined") {
      options.webSocketServer = {
        type: defaultWebSocketServerType,
        options: defaultWebSocketServerOptions,
      ***REMOVED***;
    ***REMOVED*** else if (
      typeof options.webSocketServer === "boolean" &&
      !options.webSocketServer
    ) {
      options.webSocketServer = false;
    ***REMOVED*** else if (
      typeof options.webSocketServer === "string" ||
      typeof options.webSocketServer === "function"
    ) {
      options.webSocketServer = {
        type: options.webSocketServer,
        options: defaultWebSocketServerOptions,
      ***REMOVED***;
    ***REMOVED*** else {
      options.webSocketServer = {
        type:
          /** @type {WebSocketServerConfiguration***REMOVED*** */
          (options.webSocketServer).type || defaultWebSocketServerType,
        options: {
          ...defaultWebSocketServerOptions,
          .../** @type {WebSocketServerConfiguration***REMOVED*** */
          (options.webSocketServer).options,
        ***REMOVED***,
      ***REMOVED***;

      const webSocketServer =
        /** @type {{ type: WebSocketServerConfiguration["type"], options: NonNullable<WebSocketServerConfiguration["options"]> ***REMOVED******REMOVED*** */
        (options.webSocketServer);

      if (typeof webSocketServer.options.port === "string") {
        webSocketServer.options.port = Number(webSocketServer.options.port);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  /**
   * @private
   * @returns {string***REMOVED***
   */
  getClientTransport() {
    let clientImplementation;
    let clientImplementationFound = true;

    const isKnownWebSocketServerImplementation =
      this.options.webSocketServer &&
      typeof (
        /** @type {WebSocketServerConfiguration***REMOVED*** */
        (this.options.webSocketServer).type
      ) === "string" &&
      // @ts-ignore
      (this.options.webSocketServer.type === "ws" ||
        /** @type {WebSocketServerConfiguration***REMOVED*** */
        (this.options.webSocketServer).type === "sockjs");

    let clientTransport;

    if (this.options.client) {
      if (
        typeof (
          /** @type {ClientConfiguration***REMOVED*** */
          (this.options.client).webSocketTransport
        ) !== "undefined"
      ) {
        clientTransport =
          /** @type {ClientConfiguration***REMOVED*** */
          (this.options.client).webSocketTransport;
      ***REMOVED*** else if (isKnownWebSocketServerImplementation) {
        clientTransport =
          /** @type {WebSocketServerConfiguration***REMOVED*** */
          (this.options.webSocketServer).type;
      ***REMOVED*** else {
        clientTransport = "ws";
      ***REMOVED***
    ***REMOVED*** else {
      clientTransport = "ws";
    ***REMOVED***

    switch (typeof clientTransport) {
      case "string":
        // could be 'sockjs', 'ws', or a path that should be required
        if (clientTransport === "sockjs") {
          clientImplementation = require.resolve(
            "../client/clients/SockJSClient"
          );
        ***REMOVED*** else if (clientTransport === "ws") {
          clientImplementation = require.resolve(
            "../client/clients/WebSocketClient"
          );
        ***REMOVED*** else {
          try {
            clientImplementation = require.resolve(clientTransport);
          ***REMOVED*** catch (e) {
            clientImplementationFound = false;
          ***REMOVED***
        ***REMOVED***
        break;
      default:
        clientImplementationFound = false;
    ***REMOVED***

    if (!clientImplementationFound) {
      throw new Error(
        `${
          !isKnownWebSocketServerImplementation
            ? "When you use custom web socket implementation you must explicitly specify client.webSocketTransport. "
            : ""
        ***REMOVED***client.webSocketTransport must be a string denoting a default implementation (e.g. 'sockjs', 'ws') or a full path to a JS file via require.resolve(...) which exports a class `
      );
    ***REMOVED***

    return /** @type {string***REMOVED*** */ (clientImplementation);
  ***REMOVED***

  /**
   * @private
   * @returns {string***REMOVED***
   */
  getServerTransport() {
    let implementation;
    let implementationFound = true;

    switch (
      typeof (
        /** @type {WebSocketServerConfiguration***REMOVED*** */
        (this.options.webSocketServer).type
      )
    ) {
      case "string":
        // Could be 'sockjs', in the future 'ws', or a path that should be required
        if (
          /** @type {WebSocketServerConfiguration***REMOVED*** */ (
            this.options.webSocketServer
          ).type === "sockjs"
        ) {
          implementation = require("./servers/SockJSServer");
        ***REMOVED*** else if (
          /** @type {WebSocketServerConfiguration***REMOVED*** */ (
            this.options.webSocketServer
          ).type === "ws"
        ) {
          implementation = require("./servers/WebsocketServer");
        ***REMOVED*** else {
          try {
            // eslint-disable-next-line import/no-dynamic-require
            implementation = require(/** @type {WebSocketServerConfiguration***REMOVED*** */ (
              this.options.webSocketServer
            ).type);
          ***REMOVED*** catch (error) {
            implementationFound = false;
          ***REMOVED***
        ***REMOVED***
        break;
      case "function":
        implementation = /** @type {WebSocketServerConfiguration***REMOVED*** */ (
          this.options.webSocketServer
        ).type;
        break;
      default:
        implementationFound = false;
    ***REMOVED***

    if (!implementationFound) {
      throw new Error(
        "webSocketServer (webSocketServer.type) must be a string denoting a default implementation (e.g. 'ws', 'sockjs'), a full path to " +
          "a JS file which exports a class extending BaseServer (webpack-dev-server/lib/servers/BaseServer.js) " +
          "via require.resolve(...), or the class itself which extends BaseServer"
      );
    ***REMOVED***

    return implementation;
  ***REMOVED***

  /**
   * @private
   * @returns {void***REMOVED***
   */
  setupProgressPlugin() {
    const { ProgressPlugin ***REMOVED*** =
      /** @type {MultiCompiler***REMOVED****/
      (this.compiler).compilers
        ? /** @type {MultiCompiler***REMOVED****/ (this.compiler).compilers[0].webpack
        : /** @type {Compiler***REMOVED****/ (this.compiler).webpack ||
          // TODO remove me after drop webpack v4
          require("webpack");

    new ProgressPlugin(
      /**
       * @param {number***REMOVED*** percent
       * @param {string***REMOVED*** msg
       * @param {string***REMOVED*** addInfo
       * @param {string***REMOVED*** pluginName
       */
      (percent, msg, addInfo, pluginName) => {
        percent = Math.floor(percent * 100);

        if (percent === 100) {
          msg = "Compilation completed";
        ***REMOVED***

        if (addInfo) {
          msg = `${msg***REMOVED*** (${addInfo***REMOVED***)`;
        ***REMOVED***

        if (this.webSocketServer) {
          this.sendMessage(this.webSocketServer.clients, "progress-update", {
            percent,
            msg,
            pluginName,
          ***REMOVED***);
        ***REMOVED***

        if (this.server) {
          this.server.emit("progress-update", { percent, msg, pluginName ***REMOVED***);
        ***REMOVED***
      ***REMOVED***
    ).apply(this.compiler);
  ***REMOVED***

  /**
   * @private
   * @returns {Promise<void>***REMOVED***
   */
  async initialize() {
    if (this.options.webSocketServer) {
      const compilers =
        /** @type {MultiCompiler***REMOVED*** */
        (this.compiler).compilers || [this.compiler];

      compilers.forEach((compiler) => {
        this.addAdditionalEntries(compiler);

        const webpack = compiler.webpack || require("webpack");

        new webpack.ProvidePlugin({
          __webpack_dev_server_client__: this.getClientTransport(),
        ***REMOVED***).apply(compiler);

        // TODO remove after drop webpack v4 support
        compiler.options.plugins = compiler.options.plugins || [];

        if (this.options.hot) {
          const HMRPluginExists = compiler.options.plugins.find(
            (p) => p.constructor === webpack.HotModuleReplacementPlugin
          );

          if (HMRPluginExists) {
            this.logger.warn(
              `"hot: true" automatically applies HMR plugin, you don't have to add it manually to your webpack configuration.`
            );
          ***REMOVED*** else {
            // Apply the HMR plugin
            const plugin = new webpack.HotModuleReplacementPlugin();

            plugin.apply(compiler);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***);

      if (
        this.options.client &&
        /** @type {ClientConfiguration***REMOVED*** */ (this.options.client).progress
      ) {
        this.setupProgressPlugin();
      ***REMOVED***
    ***REMOVED***

    this.setupHooks();
    this.setupApp();
    this.setupHostHeaderCheck();
    this.setupDevMiddleware();
    // Should be after `webpack-dev-middleware`, otherwise other middlewares might rewrite response
    this.setupBuiltInRoutes();
    this.setupWatchFiles();
    this.setupWatchStaticFiles();
    this.setupMiddlewares();
    this.createServer();

    if (this.options.setupExitSignals) {
      const signals = ["SIGINT", "SIGTERM"];

      let needForceShutdown = false;

      signals.forEach((signal) => {
        const listener = () => {
          if (needForceShutdown) {
            process.exit();
          ***REMOVED***

          this.logger.info(
            "Gracefully shutting down. To force exit, press ^C again. Please wait..."
          );

          needForceShutdown = true;

          this.stopCallback(() => {
            if (typeof this.compiler.close === "function") {
              this.compiler.close(() => {
                process.exit();
              ***REMOVED***);
            ***REMOVED*** else {
              process.exit();
            ***REMOVED***
          ***REMOVED***);
        ***REMOVED***;

        this.listeners.push({ name: signal, listener ***REMOVED***);

        process.on(signal, listener);
      ***REMOVED***);
    ***REMOVED***

    // Proxy WebSocket without the initial http request
    // https://github.com/chimurai/http-proxy-middleware#external-websocket-upgrade
    /** @type {RequestHandler[]***REMOVED*** */
    (this.webSocketProxies).forEach((webSocketProxy) => {
      /** @type {import("http").Server***REMOVED*** */
      (this.server).on(
        "upgrade",
        /** @type {RequestHandler & { upgrade: NonNullable<RequestHandler["upgrade"]> ***REMOVED******REMOVED*** */
        (webSocketProxy).upgrade
      );
    ***REMOVED***, this);
  ***REMOVED***

  /**
   * @private
   * @returns {void***REMOVED***
   */
  setupApp() {
    /** @type {import("express").Application | undefined***REMOVED****/
    this.app = new /** @type {any***REMOVED*** */ (getExpress())();
  ***REMOVED***

  /**
   * @private
   * @param {Stats | MultiStats***REMOVED*** statsObj
   * @returns {StatsCompilation***REMOVED***
   */
  getStats(statsObj) {
    const stats = Server.DEFAULT_STATS;
    const compilerOptions = this.getCompilerOptions();

    // @ts-ignore
    if (compilerOptions.stats && compilerOptions.stats.warningsFilter) {
      // @ts-ignore
      stats.warningsFilter = compilerOptions.stats.warningsFilter;
    ***REMOVED***

    return statsObj.toJson(stats);
  ***REMOVED***

  /**
   * @private
   * @returns {void***REMOVED***
   */
  setupHooks() {
    this.compiler.hooks.invalid.tap("webpack-dev-server", () => {
      if (this.webSocketServer) {
        this.sendMessage(this.webSocketServer.clients, "invalid");
      ***REMOVED***
    ***REMOVED***);
    this.compiler.hooks.done.tap(
      "webpack-dev-server",
      /**
       * @param {Stats | MultiStats***REMOVED*** stats
       */
      (stats) => {
        if (this.webSocketServer) {
          this.sendStats(this.webSocketServer.clients, this.getStats(stats));
        ***REMOVED***

        /**
         * @private
         * @type {Stats | MultiStats***REMOVED***
         */
        this.stats = stats;
      ***REMOVED***
    );
  ***REMOVED***

  /**
   * @private
   * @returns {void***REMOVED***
   */
  setupHostHeaderCheck() {
    /** @type {import("express").Application***REMOVED*** */
    (this.app).all(
      "*",
      /**
       * @param {Request***REMOVED*** req
       * @param {Response***REMOVED*** res
       * @param {NextFunction***REMOVED*** next
       * @returns {void***REMOVED***
       */
      (req, res, next) => {
        if (
          this.checkHeader(
            /** @type {{ [key: string]: string | undefined ***REMOVED******REMOVED*** */
            (req.headers),
            "host"
          )
        ) {
          return next();
        ***REMOVED***

        res.send("Invalid Host header");
      ***REMOVED***
    );
  ***REMOVED***

  /**
   * @private
   * @returns {void***REMOVED***
   */
  setupDevMiddleware() {
    const webpackDevMiddleware = require("webpack-dev-middleware");

    // middleware for serving webpack bundle
    this.middleware = webpackDevMiddleware(
      this.compiler,
      this.options.devMiddleware
    );
  ***REMOVED***

  /**
   * @private
   * @returns {void***REMOVED***
   */
  setupBuiltInRoutes() {
    const { app, middleware ***REMOVED*** = this;

    /** @type {import("express").Application***REMOVED*** */
    (app).get(
      "/__webpack_dev_server__/sockjs.bundle.js",
      /**
       * @param {Request***REMOVED*** req
       * @param {Response***REMOVED*** res
       * @returns {void***REMOVED***
       */
      (req, res) => {
        res.setHeader("Content-Type", "application/javascript");

        const clientPath = path.join(__dirname, "..", "client");

        res.sendFile(path.join(clientPath, "modules/sockjs-client/index.js"));
      ***REMOVED***
    );

    /** @type {import("express").Application***REMOVED*** */
    (app).get(
      "/webpack-dev-server/invalidate",
      /**
       * @param {Request***REMOVED*** _req
       * @param {Response***REMOVED*** res
       * @returns {void***REMOVED***
       */
      (_req, res) => {
        this.invalidate();

        res.end();
      ***REMOVED***
    );

    /** @type {import("express").Application***REMOVED*** */
    (app).get("/webpack-dev-server/open-editor", (req, res) => {
      const fileName = req.query.fileName;

      if (typeof fileName === "string") {
        // @ts-ignore
        const launchEditor = require("launch-editor");
        launchEditor(fileName);
      ***REMOVED***

      res.end();
    ***REMOVED***);

    /** @type {import("express").Application***REMOVED*** */
    (app).get(
      "/webpack-dev-server",
      /**
       * @param {Request***REMOVED*** req
       * @param {Response***REMOVED*** res
       * @returns {void***REMOVED***
       */
      (req, res) => {
        /** @type {import("webpack-dev-middleware").API<Request, Response>***REMOVED****/
        (middleware).waitUntilValid((stats) => {
          res.setHeader("Content-Type", "text/html");
          res.write(
            '<!DOCTYPE html><html><head><meta charset="utf-8"/></head><body>'
          );

          const statsForPrint =
            typeof (/** @type {MultiStats***REMOVED*** */ (stats).stats) !== "undefined"
              ? /** @type {MultiStats***REMOVED*** */ (stats).toJson().children
              : [/** @type {Stats***REMOVED*** */ (stats).toJson()];

          res.write(`<h1>Assets Report:</h1>`);

          /**
           * @type {StatsCompilation[]***REMOVED***
           */
          (statsForPrint).forEach((item, index) => {
            res.write("<div>");

            const name =
              // eslint-disable-next-line no-nested-ternary
              typeof item.name !== "undefined"
                ? item.name
                : /** @type {MultiStats***REMOVED*** */ (stats).stats
                ? `unnamed[${index***REMOVED***]`
                : "unnamed";

            res.write(`<h2>Compilation: ${name***REMOVED***</h2>`);
            res.write("<ul>");

            const publicPath =
              item.publicPath === "auto" ? "" : item.publicPath;

            for (const asset of /** @type {NonNullable<StatsCompilation["assets"]>***REMOVED*** */ (
              item.assets
            )) {
              const assetName = asset.name;
              const assetURL = `${publicPath***REMOVED***${assetName***REMOVED***`;

              res.write(
                `<li>
              <strong><a href="${assetURL***REMOVED***" target="_blank">${assetName***REMOVED***</a></strong>
            </li>`
              );
            ***REMOVED***

            res.write("</ul>");
            res.write("</div>");
          ***REMOVED***);

          res.end("</body></html>");
        ***REMOVED***);
      ***REMOVED***
    );
  ***REMOVED***

  /**
   * @private
   * @returns {void***REMOVED***
   */
  setupWatchStaticFiles() {
    if (/** @type {NormalizedStatic[]***REMOVED*** */ (this.options.static).length > 0) {
      /** @type {NormalizedStatic[]***REMOVED*** */
      (this.options.static).forEach((staticOption) => {
        if (staticOption.watch) {
          this.watchFiles(staticOption.directory, staticOption.watch);
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***

  /**
   * @private
   * @returns {void***REMOVED***
   */
  setupWatchFiles() {
    const { watchFiles ***REMOVED*** = this.options;

    if (/** @type {WatchFiles[]***REMOVED*** */ (watchFiles).length > 0) {
      /** @type {WatchFiles[]***REMOVED*** */
      (watchFiles).forEach((item) => {
        this.watchFiles(item.paths, item.options);
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***

  /**
   * @private
   * @returns {void***REMOVED***
   */
  setupMiddlewares() {
    /**
     * @type {Array<Middleware>***REMOVED***
     */
    let middlewares = [];

    // compress is placed last and uses unshift so that it will be the first middleware used
    if (this.options.compress) {
      const compression = require("compression");

      middlewares.push({ name: "compression", middleware: compression() ***REMOVED***);
    ***REMOVED***

    if (typeof this.options.onBeforeSetupMiddleware === "function") {
      this.options.onBeforeSetupMiddleware(this);
    ***REMOVED***

    if (typeof this.options.headers !== "undefined") {
      middlewares.push({
        name: "set-headers",
        path: "*",
        middleware: this.setHeaders.bind(this),
      ***REMOVED***);
    ***REMOVED***

    middlewares.push({
      name: "webpack-dev-middleware",
      middleware:
        /** @type {import("webpack-dev-middleware").Middleware<Request, Response>***REMOVED****/
        (this.middleware),
    ***REMOVED***);

    if (this.options.proxy) {
      const { createProxyMiddleware ***REMOVED*** = require("http-proxy-middleware");

      /**
       * @param {ProxyConfigArrayItem***REMOVED*** proxyConfig
       * @returns {RequestHandler | undefined***REMOVED***
       */
      const getProxyMiddleware = (proxyConfig) => {
        // It is possible to use the `bypass` method without a `target` or `router`.
        // However, the proxy middleware has no use in this case, and will fail to instantiate.
        if (proxyConfig.target) {
          const context = proxyConfig.context || proxyConfig.path;

          return createProxyMiddleware(
            /** @type {string***REMOVED*** */ (context),
            proxyConfig
          );
        ***REMOVED***

        if (proxyConfig.router) {
          return createProxyMiddleware(proxyConfig);
        ***REMOVED***
      ***REMOVED***;

      /**
       * Assume a proxy configuration specified as:
       * proxy: [
       *   {
       *     context: "value",
       *     ...options,
       *   ***REMOVED***,
       *   // or:
       *   function() {
       *     return {
       *       context: "context",
       *       ...options,
       *     ***REMOVED***;
       *   ***REMOVED***
       * ]
       */
      /** @type {ProxyConfigArray***REMOVED*** */
      (this.options.proxy).forEach((proxyConfigOrCallback) => {
        /**
         * @type {RequestHandler***REMOVED***
         */
        let proxyMiddleware;

        let proxyConfig =
          typeof proxyConfigOrCallback === "function"
            ? proxyConfigOrCallback()
            : proxyConfigOrCallback;

        proxyMiddleware =
          /** @type {RequestHandler***REMOVED*** */
          (getProxyMiddleware(proxyConfig));

        if (proxyConfig.ws) {
          this.webSocketProxies.push(proxyMiddleware);
        ***REMOVED***

        /**
         * @param {Request***REMOVED*** req
         * @param {Response***REMOVED*** res
         * @param {NextFunction***REMOVED*** next
         * @returns {Promise<void>***REMOVED***
         */
        const handler = async (req, res, next) => {
          if (typeof proxyConfigOrCallback === "function") {
            const newProxyConfig = proxyConfigOrCallback(req, res, next);

            if (newProxyConfig !== proxyConfig) {
              proxyConfig = newProxyConfig;
              proxyMiddleware =
                /** @type {RequestHandler***REMOVED*** */
                (getProxyMiddleware(proxyConfig));
            ***REMOVED***
          ***REMOVED***

          // - Check if we have a bypass function defined
          // - In case the bypass function is defined we'll retrieve the
          // bypassUrl from it otherwise bypassUrl would be null
          // TODO remove in the next major in favor `context` and `router` options
          const isByPassFuncDefined = typeof proxyConfig.bypass === "function";
          const bypassUrl = isByPassFuncDefined
            ? await /** @type {ByPass***REMOVED*** */ (proxyConfig.bypass)(
                req,
                res,
                proxyConfig
              )
            : null;

          if (typeof bypassUrl === "boolean") {
            // skip the proxy
            // @ts-ignore
            req.url = null;
            next();
          ***REMOVED*** else if (typeof bypassUrl === "string") {
            // byPass to that url
            req.url = bypassUrl;
            next();
          ***REMOVED*** else if (proxyMiddleware) {
            return proxyMiddleware(req, res, next);
          ***REMOVED*** else {
            next();
          ***REMOVED***
        ***REMOVED***;

        middlewares.push({
          name: "http-proxy-middleware",
          middleware: handler,
        ***REMOVED***);
        // Also forward error requests to the proxy so it can handle them.
        middlewares.push({
          name: "http-proxy-middleware-error-handler",
          middleware:
            /**
             * @param {Error***REMOVED*** error
             * @param {Request***REMOVED*** req
             * @param {Response***REMOVED*** res
             * @param {NextFunction***REMOVED*** next
             * @returns {any***REMOVED***
             */
            (error, req, res, next) => handler(req, res, next),
        ***REMOVED***);
      ***REMOVED***);

      middlewares.push({
        name: "webpack-dev-middleware",
        middleware:
          /** @type {import("webpack-dev-middleware").Middleware<Request, Response>***REMOVED****/
          (this.middleware),
      ***REMOVED***);
    ***REMOVED***

    if (/** @type {NormalizedStatic[]***REMOVED*** */ (this.options.static).length > 0) {
      /** @type {NormalizedStatic[]***REMOVED*** */
      (this.options.static).forEach((staticOption) => {
        staticOption.publicPath.forEach((publicPath) => {
          middlewares.push({
            name: "express-static",
            path: publicPath,
            middleware: getExpress().static(
              staticOption.directory,
              staticOption.staticOptions
            ),
          ***REMOVED***);
        ***REMOVED***);
      ***REMOVED***);
    ***REMOVED***

    if (this.options.historyApiFallback) {
      const connectHistoryApiFallback = require("connect-history-api-fallback");
      const { historyApiFallback ***REMOVED*** = this.options;

      if (
        typeof (
          /** @type {ConnectHistoryApiFallbackOptions***REMOVED*** */
          (historyApiFallback).logger
        ) === "undefined" &&
        !(
          /** @type {ConnectHistoryApiFallbackOptions***REMOVED*** */
          (historyApiFallback).verbose
        )
      ) {
        // @ts-ignore
        historyApiFallback.logger = this.logger.log.bind(
          this.logger,
          "[connect-history-api-fallback]"
        );
      ***REMOVED***

      // Fall back to /index.html if nothing else matches.
      middlewares.push({
        name: "connect-history-api-fallback",
        middleware: connectHistoryApiFallback(
          /** @type {ConnectHistoryApiFallbackOptions***REMOVED*** */
          (historyApiFallback)
        ),
      ***REMOVED***);

      // include our middleware to ensure
      // it is able to handle '/index.html' request after redirect
      middlewares.push({
        name: "webpack-dev-middleware",
        middleware:
          /** @type {import("webpack-dev-middleware").Middleware<Request, Response>***REMOVED****/
          (this.middleware),
      ***REMOVED***);

      if (/** @type {NormalizedStatic[]***REMOVED*** */ (this.options.static).length > 0) {
        /** @type {NormalizedStatic[]***REMOVED*** */
        (this.options.static).forEach((staticOption) => {
          staticOption.publicPath.forEach((publicPath) => {
            middlewares.push({
              name: "express-static",
              path: publicPath,
              middleware: getExpress().static(
                staticOption.directory,
                staticOption.staticOptions
              ),
            ***REMOVED***);
          ***REMOVED***);
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***

    if (/** @type {NormalizedStatic[]***REMOVED*** */ (this.options.static).length > 0) {
      const serveIndex = require("serve-index");

      /** @type {NormalizedStatic[]***REMOVED*** */
      (this.options.static).forEach((staticOption) => {
        staticOption.publicPath.forEach((publicPath) => {
          if (staticOption.serveIndex) {
            middlewares.push({
              name: "serve-index",
              path: publicPath,
              /**
               * @param {Request***REMOVED*** req
               * @param {Response***REMOVED*** res
               * @param {NextFunction***REMOVED*** next
               * @returns {void***REMOVED***
               */
              middleware: (req, res, next) => {
                // serve-index doesn't fallthrough non-get/head request to next middleware
                if (req.method !== "GET" && req.method !== "HEAD") {
                  return next();
                ***REMOVED***

                serveIndex(
                  staticOption.directory,
                  /** @type {ServeIndexOptions***REMOVED*** */
                  (staticOption.serveIndex)
                )(req, res, next);
              ***REMOVED***,
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED***);
      ***REMOVED***);
    ***REMOVED***

    if (this.options.magicHtml) {
      middlewares.push({
        name: "serve-magic-html",
        middleware: this.serveMagicHtml.bind(this),
      ***REMOVED***);
    ***REMOVED***

    // Register this middleware always as the last one so that it's only used as a
    // fallback when no other middleware responses.
    middlewares.push({
      name: "options-middleware",
      path: "*",
      /**
       * @param {Request***REMOVED*** req
       * @param {Response***REMOVED*** res
       * @param {NextFunction***REMOVED*** next
       * @returns {void***REMOVED***
       */
      middleware: (req, res, next) => {
        if (req.method === "OPTIONS") {
          res.statusCode = 204;
          res.setHeader("Content-Length", "0");
          res.end();
          return;
        ***REMOVED***
        next();
      ***REMOVED***,
    ***REMOVED***);

    if (typeof this.options.setupMiddlewares === "function") {
      middlewares = this.options.setupMiddlewares(middlewares, this);
    ***REMOVED***

    middlewares.forEach((middleware) => {
      if (typeof middleware === "function") {
        /** @type {import("express").Application***REMOVED*** */
        (this.app).use(middleware);
      ***REMOVED*** else if (typeof middleware.path !== "undefined") {
        /** @type {import("express").Application***REMOVED*** */
        (this.app).use(middleware.path, middleware.middleware);
      ***REMOVED*** else {
        /** @type {import("express").Application***REMOVED*** */
        (this.app).use(middleware.middleware);
      ***REMOVED***
    ***REMOVED***);

    if (typeof this.options.onAfterSetupMiddleware === "function") {
      this.options.onAfterSetupMiddleware(this);
    ***REMOVED***
  ***REMOVED***

  /**
   * @private
   * @returns {void***REMOVED***
   */
  createServer() {
    const { type, options ***REMOVED*** = /** @type {ServerConfiguration***REMOVED*** */ (
      this.options.server
    );

    /** @type {import("http").Server | undefined | null***REMOVED*** */
    // eslint-disable-next-line import/no-dynamic-require
    this.server = require(/** @type {string***REMOVED*** */ (type)).createServer(
      options,
      this.app
    );

    /** @type {import("http").Server***REMOVED*** */
    (this.server).on(
      "connection",
      /**
       * @param {Socket***REMOVED*** socket
       */
      (socket) => {
        // Add socket to list
        this.sockets.push(socket);

        socket.once("close", () => {
          // Remove socket from list
          this.sockets.splice(this.sockets.indexOf(socket), 1);
        ***REMOVED***);
      ***REMOVED***
    );

    /** @type {import("http").Server***REMOVED*** */
    (this.server).on(
      "error",
      /**
       * @param {Error***REMOVED*** error
       */
      (error) => {
        throw error;
      ***REMOVED***
    );
  ***REMOVED***

  /**
   * @private
   * @returns {void***REMOVED***
   */
  // TODO: remove `--web-socket-server` in favor of `--web-socket-server-type`
  createWebSocketServer() {
    /** @type {WebSocketServerImplementation | undefined | null***REMOVED*** */
    this.webSocketServer = new /** @type {any***REMOVED*** */ (this.getServerTransport())(
      this
    );
    /** @type {WebSocketServerImplementation***REMOVED*** */
    (this.webSocketServer).implementation.on(
      "connection",
      /**
       * @param {ClientConnection***REMOVED*** client
       * @param {IncomingMessage***REMOVED*** request
       */
      (client, request) => {
        /** @type {{ [key: string]: string | undefined ***REMOVED*** | undefined***REMOVED*** */
        const headers =
          // eslint-disable-next-line no-nested-ternary
          typeof request !== "undefined"
            ? /** @type {{ [key: string]: string | undefined ***REMOVED******REMOVED*** */
              (request.headers)
            : typeof (
                /** @type {import("sockjs").Connection***REMOVED*** */ (client).headers
              ) !== "undefined"
            ? /** @type {import("sockjs").Connection***REMOVED*** */ (client).headers
            : // eslint-disable-next-line no-undefined
              undefined;

        if (!headers) {
          this.logger.warn(
            'webSocketServer implementation must pass headers for the "connection" event'
          );
        ***REMOVED***

        if (
          !headers ||
          !this.checkHeader(headers, "host") ||
          !this.checkHeader(headers, "origin")
        ) {
          this.sendMessage([client], "error", "Invalid Host/Origin header");

          // With https enabled, the sendMessage above is encrypted asynchronously so not yet sent
          // Terminate would prevent it sending, so use close to allow it to be sent
          client.close();

          return;
        ***REMOVED***

        if (this.options.hot === true || this.options.hot === "only") {
          this.sendMessage([client], "hot");
        ***REMOVED***

        if (this.options.liveReload) {
          this.sendMessage([client], "liveReload");
        ***REMOVED***

        if (
          this.options.client &&
          /** @type {ClientConfiguration***REMOVED*** */
          (this.options.client).progress
        ) {
          this.sendMessage(
            [client],
            "progress",
            /** @type {ClientConfiguration***REMOVED*** */
            (this.options.client).progress
          );
        ***REMOVED***

        if (
          this.options.client &&
          /** @type {ClientConfiguration***REMOVED*** */ (this.options.client).reconnect
        ) {
          this.sendMessage(
            [client],
            "reconnect",
            /** @type {ClientConfiguration***REMOVED*** */
            (this.options.client).reconnect
          );
        ***REMOVED***

        if (
          this.options.client &&
          /** @type {ClientConfiguration***REMOVED*** */
          (this.options.client).overlay
        ) {
          const overlayConfig = /** @type {ClientConfiguration***REMOVED*** */ (
            this.options.client
          ).overlay;

          this.sendMessage(
            [client],
            "overlay",
            typeof overlayConfig === "object"
              ? {
                  ...overlayConfig,
                  errors:
                    overlayConfig.errors &&
                    encodeOverlaySettings(overlayConfig.errors),
                  warnings:
                    overlayConfig.warnings &&
                    encodeOverlaySettings(overlayConfig.warnings),
                  runtimeErrors:
                    overlayConfig.runtimeErrors &&
                    encodeOverlaySettings(overlayConfig.runtimeErrors),
                ***REMOVED***
              : overlayConfig
          );
        ***REMOVED***

        if (!this.stats) {
          return;
        ***REMOVED***

        this.sendStats([client], this.getStats(this.stats), true);
      ***REMOVED***
    );
  ***REMOVED***

  /**
   * @private
   * @param {string***REMOVED*** defaultOpenTarget
   * @returns {void***REMOVED***
   */
  openBrowser(defaultOpenTarget) {
    const open = require("open");

    Promise.all(
      /** @type {NormalizedOpen[]***REMOVED*** */
      (this.options.open).map((item) => {
        /**
         * @type {string***REMOVED***
         */
        let openTarget;

        if (item.target === "<url>") {
          openTarget = defaultOpenTarget;
        ***REMOVED*** else {
          openTarget = Server.isAbsoluteURL(item.target)
            ? item.target
            : new URL(item.target, defaultOpenTarget).toString();
        ***REMOVED***

        return open(openTarget, item.options).catch(() => {
          this.logger.warn(
            `Unable to open "${openTarget***REMOVED***" page${
              item.options.app
                ? ` in "${
                    /** @type {import("open").App***REMOVED*** */
                    (item.options.app).name
                  ***REMOVED***" app${
                    /** @type {import("open").App***REMOVED*** */
                    (item.options.app).arguments
                      ? ` with "${
                          /** @type {import("open").App***REMOVED*** */
                          (item.options.app).arguments.join(" ")
                        ***REMOVED***" arguments`
                      : ""
                  ***REMOVED***`
                : ""
            ***REMOVED***. If you are running in a headless environment, please do not use the "open" option or related flags like "--open", "--open-target", and "--open-app".`
          );
        ***REMOVED***);
      ***REMOVED***)
    );
  ***REMOVED***

  /**
   * @private
   * @returns {void***REMOVED***
   */
  runBonjour() {
    const { Bonjour ***REMOVED*** = require("bonjour-service");
    /**
     * @private
     * @type {Bonjour | undefined***REMOVED***
     */
    this.bonjour = new Bonjour();
    this.bonjour.publish({
      // @ts-expect-error
      name: `Webpack Dev Server ${os.hostname()***REMOVED***:${this.options.port***REMOVED***`,
      // @ts-expect-error
      port: /** @type {number***REMOVED*** */ (this.options.port),
      // @ts-expect-error
      type:
        /** @type {ServerConfiguration***REMOVED*** */
        (this.options.server).type === "http" ? "http" : "https",
      subtypes: ["webpack"],
      .../** @type {BonjourOptions***REMOVED*** */ (this.options.bonjour),
    ***REMOVED***);
  ***REMOVED***

  /**
   * @private
   * @returns {void***REMOVED***
   */
  stopBonjour(callback = () => {***REMOVED***) {
    /** @type {Bonjour***REMOVED*** */
    (this.bonjour).unpublishAll(() => {
      /** @type {Bonjour***REMOVED*** */
      (this.bonjour).destroy();

      if (callback) {
        callback();
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***

  /**
   * @private
   * @returns {void***REMOVED***
   */
  logStatus() {
    const { isColorSupported, cyan, red ***REMOVED*** = require("colorette");

    /**
     * @param {Compiler["options"]***REMOVED*** compilerOptions
     * @returns {boolean***REMOVED***
     */
    const getColorsOption = (compilerOptions) => {
      /**
       * @type {boolean***REMOVED***
       */
      let colorsEnabled;

      if (
        compilerOptions.stats &&
        typeof (/** @type {StatsOptions***REMOVED*** */ (compilerOptions.stats).colors) !==
          "undefined"
      ) {
        colorsEnabled =
          /** @type {boolean***REMOVED*** */
          (/** @type {StatsOptions***REMOVED*** */ (compilerOptions.stats).colors);
      ***REMOVED*** else {
        colorsEnabled = isColorSupported;
      ***REMOVED***

      return colorsEnabled;
    ***REMOVED***;

    const colors = {
      /**
       * @param {boolean***REMOVED*** useColor
       * @param {string***REMOVED*** msg
       * @returns {string***REMOVED***
       */
      info(useColor, msg) {
        if (useColor) {
          return cyan(msg);
        ***REMOVED***

        return msg;
      ***REMOVED***,
      /**
       * @param {boolean***REMOVED*** useColor
       * @param {string***REMOVED*** msg
       * @returns {string***REMOVED***
       */
      error(useColor, msg) {
        if (useColor) {
          return red(msg);
        ***REMOVED***

        return msg;
      ***REMOVED***,
    ***REMOVED***;
    const useColor = getColorsOption(this.getCompilerOptions());

    if (this.options.ipc) {
      this.logger.info(
        `Project is running at: "${
          /** @type {import("http").Server***REMOVED*** */
          (this.server).address()
        ***REMOVED***"`
      );
    ***REMOVED*** else {
      const protocol =
        /** @type {ServerConfiguration***REMOVED*** */
        (this.options.server).type === "http" ? "http" : "https";
      const { address, port ***REMOVED*** =
        /** @type {import("net").AddressInfo***REMOVED*** */
        (
          /** @type {import("http").Server***REMOVED*** */
          (this.server).address()
        );
      /**
       * @param {string***REMOVED*** newHostname
       * @returns {string***REMOVED***
       */
      const prettyPrintURL = (newHostname) =>
        url.format({ protocol, hostname: newHostname, port, pathname: "/" ***REMOVED***);

      let server;
      let localhost;
      let loopbackIPv4;
      let loopbackIPv6;
      let networkUrlIPv4;
      let networkUrlIPv6;

      if (this.options.host) {
        if (this.options.host === "localhost") {
          localhost = prettyPrintURL("localhost");
        ***REMOVED*** else {
          let isIP;

          try {
            isIP = ipaddr.parse(this.options.host);
          ***REMOVED*** catch (error) {
            // Ignore
          ***REMOVED***

          if (!isIP) {
            server = prettyPrintURL(this.options.host);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***

      const parsedIP = ipaddr.parse(address);

      if (parsedIP.range() === "unspecified") {
        localhost = prettyPrintURL("localhost");

        const networkIPv4 = Server.internalIPSync("v4");

        if (networkIPv4) {
          networkUrlIPv4 = prettyPrintURL(networkIPv4);
        ***REMOVED***

        const networkIPv6 = Server.internalIPSync("v6");

        if (networkIPv6) {
          networkUrlIPv6 = prettyPrintURL(networkIPv6);
        ***REMOVED***
      ***REMOVED*** else if (parsedIP.range() === "loopback") {
        if (parsedIP.kind() === "ipv4") {
          loopbackIPv4 = prettyPrintURL(parsedIP.toString());
        ***REMOVED*** else if (parsedIP.kind() === "ipv6") {
          loopbackIPv6 = prettyPrintURL(parsedIP.toString());
        ***REMOVED***
      ***REMOVED*** else {
        networkUrlIPv4 =
          parsedIP.kind() === "ipv6" &&
          /** @type {IPv6***REMOVED*** */
          (parsedIP).isIPv4MappedAddress()
            ? prettyPrintURL(
                /** @type {IPv6***REMOVED*** */
                (parsedIP).toIPv4Address().toString()
              )
            : prettyPrintURL(address);

        if (parsedIP.kind() === "ipv6") {
          networkUrlIPv6 = prettyPrintURL(address);
        ***REMOVED***
      ***REMOVED***

      this.logger.info("Project is running at:");

      if (server) {
        this.logger.info(`Server: ${colors.info(useColor, server)***REMOVED***`);
      ***REMOVED***

      if (localhost || loopbackIPv4 || loopbackIPv6) {
        const loopbacks = [];

        if (localhost) {
          loopbacks.push([colors.info(useColor, localhost)]);
        ***REMOVED***

        if (loopbackIPv4) {
          loopbacks.push([colors.info(useColor, loopbackIPv4)]);
        ***REMOVED***

        if (loopbackIPv6) {
          loopbacks.push([colors.info(useColor, loopbackIPv6)]);
        ***REMOVED***

        this.logger.info(`Loopback: ${loopbacks.join(", ")***REMOVED***`);
      ***REMOVED***

      if (networkUrlIPv4) {
        this.logger.info(
          `On Your Network (IPv4): ${colors.info(useColor, networkUrlIPv4)***REMOVED***`
        );
      ***REMOVED***

      if (networkUrlIPv6) {
        this.logger.info(
          `On Your Network (IPv6): ${colors.info(useColor, networkUrlIPv6)***REMOVED***`
        );
      ***REMOVED***

      if (/** @type {NormalizedOpen[]***REMOVED*** */ (this.options.open).length > 0) {
        const openTarget = prettyPrintURL(
          !this.options.host ||
            this.options.host === "0.0.0.0" ||
            this.options.host === "::"
            ? "localhost"
            : this.options.host
        );

        this.openBrowser(openTarget);
      ***REMOVED***
    ***REMOVED***

    if (/** @type {NormalizedStatic[]***REMOVED*** */ (this.options.static).length > 0) {
      this.logger.info(
        `Content not from webpack is served from '${colors.info(
          useColor,
          /** @type {NormalizedStatic[]***REMOVED*** */
          (this.options.static)
            .map((staticOption) => staticOption.directory)
            .join(", ")
        )***REMOVED***' directory`
      );
    ***REMOVED***

    if (this.options.historyApiFallback) {
      this.logger.info(
        `404s will fallback to '${colors.info(
          useColor,
          /** @type {ConnectHistoryApiFallbackOptions***REMOVED*** */ (
            this.options.historyApiFallback
          ).index || "/index.html"
        )***REMOVED***'`
      );
    ***REMOVED***

    if (this.options.bonjour) {
      const bonjourProtocol =
        /** @type {BonjourOptions***REMOVED*** */
        (this.options.bonjour).type ||
        /** @type {ServerConfiguration***REMOVED*** */
        (this.options.server).type === "http"
          ? "http"
          : "https";

      this.logger.info(
        `Broadcasting "${bonjourProtocol***REMOVED***" with subtype of "webpack" via ZeroConf DNS (Bonjour)`
      );
    ***REMOVED***
  ***REMOVED***

  /**
   * @private
   * @param {Request***REMOVED*** req
   * @param {Response***REMOVED*** res
   * @param {NextFunction***REMOVED*** next
   */
  setHeaders(req, res, next) {
    let { headers ***REMOVED*** = this.options;

    if (headers) {
      if (typeof headers === "function") {
        headers = headers(
          req,
          res,
          /** @type {import("webpack-dev-middleware").API<Request, Response>***REMOVED****/
          (this.middleware).context
        );
      ***REMOVED***

      /**
       * @type {{key: string, value: string***REMOVED***[]***REMOVED***
       */
      const allHeaders = [];

      if (!Array.isArray(headers)) {
        // eslint-disable-next-line guard-for-in
        for (const name in headers) {
          // @ts-ignore
          allHeaders.push({ key: name, value: headers[name] ***REMOVED***);
        ***REMOVED***

        headers = allHeaders;
      ***REMOVED***

      headers.forEach(
        /**
         * @param {{key: string, value: any***REMOVED******REMOVED*** header
         */
        (header) => {
          res.setHeader(header.key, header.value);
        ***REMOVED***
      );
    ***REMOVED***

    next();
  ***REMOVED***

  /**
   * @private
   * @param {{ [key: string]: string | undefined ***REMOVED******REMOVED*** headers
   * @param {string***REMOVED*** headerToCheck
   * @returns {boolean***REMOVED***
   */
  checkHeader(headers, headerToCheck) {
    // allow user to opt out of this security check, at their own risk
    // by explicitly enabling allowedHosts
    if (this.options.allowedHosts === "all") {
      return true;
    ***REMOVED***

    // get the Host header and extract hostname
    // we don't care about port not matching
    const hostHeader = headers[headerToCheck];

    if (!hostHeader) {
      return false;
    ***REMOVED***

    if (/^(file|.+-extension):/i.test(hostHeader)) {
      return true;
    ***REMOVED***

    // use the node url-parser to retrieve the hostname from the host-header.
    const hostname = url.parse(
      // if hostHeader doesn't have scheme, add // for parsing.
      /^(.+:)?\/\//.test(hostHeader) ? hostHeader : `//${hostHeader***REMOVED***`,
      false,
      true
    ).hostname;

    // always allow requests with explicit IPv4 or IPv6-address.
    // A note on IPv6 addresses:
    // hostHeader will always contain the brackets denoting
    // an IPv6-address in URLs,
    // these are removed from the hostname in url.parse(),
    // so we have the pure IPv6-address in hostname.
    // For convenience, always allow localhost (hostname === 'localhost')
    // and its subdomains (hostname.endsWith(".localhost")).
    // allow hostname of listening address  (hostname === this.options.host)
    const isValidHostname =
      (hostname !== null && ipaddr.IPv4.isValid(hostname)) ||
      (hostname !== null && ipaddr.IPv6.isValid(hostname)) ||
      hostname === "localhost" ||
      (hostname !== null && hostname.endsWith(".localhost")) ||
      hostname === this.options.host;

    if (isValidHostname) {
      return true;
    ***REMOVED***

    const { allowedHosts ***REMOVED*** = this.options;

    // always allow localhost host, for convenience
    // allow if hostname is in allowedHosts
    if (Array.isArray(allowedHosts) && allowedHosts.length > 0) {
      for (let hostIdx = 0; hostIdx < allowedHosts.length; hostIdx++) {
        const allowedHost = allowedHosts[hostIdx];

        if (allowedHost === hostname) {
          return true;
        ***REMOVED***

        // support "." as a subdomain wildcard
        // e.g. ".example.com" will allow "example.com", "www.example.com", "subdomain.example.com", etc
        if (allowedHost[0] === ".") {
          // "example.com"  (hostname === allowedHost.substring(1))
          // "*.example.com"  (hostname.endsWith(allowedHost))
          if (
            hostname === allowedHost.substring(1) ||
            /** @type {string***REMOVED*** */ (hostname).endsWith(allowedHost)
          ) {
            return true;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    // Also allow if `client.webSocketURL.hostname` provided
    if (
      this.options.client &&
      typeof (
        /** @type {ClientConfiguration***REMOVED*** */ (this.options.client).webSocketURL
      ) !== "undefined"
    ) {
      return (
        /** @type {WebSocketURL***REMOVED*** */
        (/** @type {ClientConfiguration***REMOVED*** */ (this.options.client).webSocketURL)
          .hostname === hostname
      );
    ***REMOVED***

    // disallow
    return false;
  ***REMOVED***

  /**
   * @param {ClientConnection[]***REMOVED*** clients
   * @param {string***REMOVED*** type
   * @param {any***REMOVED*** [data]
   * @param {any***REMOVED*** [params]
   */
  // eslint-disable-next-line class-methods-use-this
  sendMessage(clients, type, data, params) {
    for (const client of clients) {
      // `sockjs` uses `1` to indicate client is ready to accept data
      // `ws` uses `WebSocket.OPEN`, but it is mean `1` too
      if (client.readyState === 1) {
        client.send(JSON.stringify({ type, data, params ***REMOVED***));
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  /**
   * @private
   * @param {Request***REMOVED*** req
   * @param {Response***REMOVED*** res
   * @param {NextFunction***REMOVED*** next
   * @returns {void***REMOVED***
   */
  serveMagicHtml(req, res, next) {
    if (req.method !== "GET" && req.method !== "HEAD") {
      return next();
    ***REMOVED***

    /** @type {import("webpack-dev-middleware").API<Request, Response>***REMOVED****/
    (this.middleware).waitUntilValid(() => {
      const _path = req.path;

      try {
        const filename =
          /** @type {import("webpack-dev-middleware").API<Request, Response>***REMOVED****/
          (this.middleware).getFilenameFromUrl(`${_path***REMOVED***.js`);
        const isFile =
          /** @type {Compiler["outputFileSystem"] & { statSync: import("fs").StatSyncFn ***REMOVED******REMOVED****/
          (
            /** @type {import("webpack-dev-middleware").API<Request, Response>***REMOVED****/
            (this.middleware).context.outputFileSystem
          )
            .statSync(/** @type {import("fs").PathLike***REMOVED*** */ (filename))
            .isFile();

        if (!isFile) {
          return next();
        ***REMOVED***

        // Serve a page that executes the javascript
        // @ts-ignore
        const queries = req._parsedUrl.search || "";
        const responsePage = `<!DOCTYPE html><html><head><meta charset="utf-8"/></head><body><script type="text/javascript" charset="utf-8" src="${_path***REMOVED***.js${queries***REMOVED***"></script></body></html>`;

        res.send(responsePage);
      ***REMOVED*** catch (error) {
        return next();
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***

  // Send stats to a socket or multiple sockets
  /**
   * @private
   * @param {ClientConnection[]***REMOVED*** clients
   * @param {StatsCompilation***REMOVED*** stats
   * @param {boolean***REMOVED*** [force]
   */
  sendStats(clients, stats, force) {
    const shouldEmit =
      !force &&
      stats &&
      (!stats.errors || stats.errors.length === 0) &&
      (!stats.warnings || stats.warnings.length === 0) &&
      this.currentHash === stats.hash;

    if (shouldEmit) {
      this.sendMessage(clients, "still-ok");

      return;
    ***REMOVED***

    this.currentHash = stats.hash;
    this.sendMessage(clients, "hash", stats.hash);

    if (
      /** @type {NonNullable<StatsCompilation["errors"]>***REMOVED*** */
      (stats.errors).length > 0 ||
      /** @type {NonNullable<StatsCompilation["warnings"]>***REMOVED*** */
      (stats.warnings).length > 0
    ) {
      const hasErrors =
        /** @type {NonNullable<StatsCompilation["errors"]>***REMOVED*** */
        (stats.errors).length > 0;

      if (
        /** @type {NonNullable<StatsCompilation["warnings"]>***REMOVED*** */
        (stats.warnings).length > 0
      ) {
        let params;

        if (hasErrors) {
          params = { preventReloading: true ***REMOVED***;
        ***REMOVED***

        this.sendMessage(clients, "warnings", stats.warnings, params);
      ***REMOVED***

      if (
        /** @type {NonNullable<StatsCompilation["errors"]>***REMOVED*** */ (stats.errors)
          .length > 0
      ) {
        this.sendMessage(clients, "errors", stats.errors);
      ***REMOVED***
    ***REMOVED*** else {
      this.sendMessage(clients, "ok");
    ***REMOVED***
  ***REMOVED***

  /**
   * @param {string | string[]***REMOVED*** watchPath
   * @param {WatchOptions***REMOVED*** [watchOptions]
   */
  watchFiles(watchPath, watchOptions) {
    const chokidar = require("chokidar");
    const watcher = chokidar.watch(watchPath, watchOptions);

    // disabling refreshing on changing the content
    if (this.options.liveReload) {
      watcher.on("change", (item) => {
        if (this.webSocketServer) {
          this.sendMessage(
            this.webSocketServer.clients,
            "static-changed",
            item
          );
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***

    this.staticWatchers.push(watcher);
  ***REMOVED***

  /**
   * @param {import("webpack-dev-middleware").Callback***REMOVED*** [callback]
   */
  invalidate(callback = () => {***REMOVED***) {
    if (this.middleware) {
      this.middleware.invalidate(callback);
    ***REMOVED***
  ***REMOVED***

  /**
   * @returns {Promise<void>***REMOVED***
   */
  async start() {
    await this.normalizeOptions();

    if (this.options.ipc) {
      await /** @type {Promise<void>***REMOVED*** */ (
        new Promise((resolve, reject) => {
          const net = require("net");
          const socket = new net.Socket();

          socket.on(
            "error",
            /**
             * @param {Error & { code?: string ***REMOVED******REMOVED*** error
             */
            (error) => {
              if (error.code === "ECONNREFUSED") {
                // No other server listening on this socket, so it can be safely removed
                fs.unlinkSync(/** @type {string***REMOVED*** */ (this.options.ipc));

                resolve();

                return;
              ***REMOVED*** else if (error.code === "ENOENT") {
                resolve();

                return;
              ***REMOVED***

              reject(error);
            ***REMOVED***
          );

          socket.connect(
            { path: /** @type {string***REMOVED*** */ (this.options.ipc) ***REMOVED***,
            () => {
              throw new Error(`IPC "${this.options.ipc***REMOVED***" is already used`);
            ***REMOVED***
          );
        ***REMOVED***)
      );
    ***REMOVED*** else {
      this.options.host = await Server.getHostname(
        /** @type {Host***REMOVED*** */ (this.options.host)
      );
      this.options.port = await Server.getFreePort(
        /** @type {Port***REMOVED*** */ (this.options.port),
        this.options.host
      );
    ***REMOVED***

    await this.initialize();

    const listenOptions = this.options.ipc
      ? { path: this.options.ipc ***REMOVED***
      : { host: this.options.host, port: this.options.port ***REMOVED***;

    await /** @type {Promise<void>***REMOVED*** */ (
      new Promise((resolve) => {
        /** @type {import("http").Server***REMOVED*** */
        (this.server).listen(listenOptions, () => {
          resolve();
        ***REMOVED***);
      ***REMOVED***)
    );

    if (this.options.ipc) {
      // chmod 666 (rw rw rw)
      const READ_WRITE = 438;

      await fs.promises.chmod(
        /** @type {string***REMOVED*** */ (this.options.ipc),
        READ_WRITE
      );
    ***REMOVED***

    if (this.options.webSocketServer) {
      this.createWebSocketServer();
    ***REMOVED***

    if (this.options.bonjour) {
      this.runBonjour();
    ***REMOVED***

    this.logStatus();

    if (typeof this.options.onListening === "function") {
      this.options.onListening(this);
    ***REMOVED***
  ***REMOVED***

  /**
   * @param {(err?: Error) => void***REMOVED*** [callback]
   */
  startCallback(callback = () => {***REMOVED***) {
    this.start()
      .then(() => callback(), callback)
      .catch(callback);
  ***REMOVED***

  /**
   * @returns {Promise<void>***REMOVED***
   */
  async stop() {
    if (this.bonjour) {
      await /** @type {Promise<void>***REMOVED*** */ (
        new Promise((resolve) => {
          this.stopBonjour(() => {
            resolve();
          ***REMOVED***);
        ***REMOVED***)
      );
    ***REMOVED***

    this.webSocketProxies = [];

    await Promise.all(this.staticWatchers.map((watcher) => watcher.close()));

    this.staticWatchers = [];

    if (this.webSocketServer) {
      await /** @type {Promise<void>***REMOVED*** */ (
        new Promise((resolve) => {
          /** @type {WebSocketServerImplementation***REMOVED*** */
          (this.webSocketServer).implementation.close(() => {
            this.webSocketServer = null;

            resolve();
          ***REMOVED***);

          for (const client of /** @type {WebSocketServerImplementation***REMOVED*** */ (
            this.webSocketServer
          ).clients) {
            client.terminate();
          ***REMOVED***

          /** @type {WebSocketServerImplementation***REMOVED*** */
          (this.webSocketServer).clients = [];
        ***REMOVED***)
      );
    ***REMOVED***

    if (this.server) {
      await /** @type {Promise<void>***REMOVED*** */ (
        new Promise((resolve) => {
          /** @type {import("http").Server***REMOVED*** */
          (this.server).close(() => {
            this.server = null;

            resolve();
          ***REMOVED***);

          for (const socket of this.sockets) {
            socket.destroy();
          ***REMOVED***

          this.sockets = [];
        ***REMOVED***)
      );

      if (this.middleware) {
        await /** @type {Promise<void>***REMOVED*** */ (
          new Promise((resolve, reject) => {
            /** @type {import("webpack-dev-middleware").API<Request, Response>***REMOVED****/
            (this.middleware).close((error) => {
              if (error) {
                reject(error);

                return;
              ***REMOVED***

              resolve();
            ***REMOVED***);
          ***REMOVED***)
        );

        this.middleware = null;
      ***REMOVED***
    ***REMOVED***

    // We add listeners to signals when creating a new Server instance
    // So ensure they are removed to prevent EventEmitter memory leak warnings
    for (const item of this.listeners) {
      process.removeListener(item.name, item.listener);
    ***REMOVED***
  ***REMOVED***

  /**
   * @param {(err?: Error) => void***REMOVED*** [callback]
   */
  stopCallback(callback = () => {***REMOVED***) {
    this.stop()
      .then(() => callback(), callback)
      .catch(callback);
  ***REMOVED***

  // TODO remove in the next major release
  /**
   * @param {Port***REMOVED*** port
   * @param {Host***REMOVED*** hostname
   * @param {(err?: Error) => void***REMOVED*** fn
   * @returns {void***REMOVED***
   */
  listen(port, hostname, fn) {
    util.deprecate(
      () => {***REMOVED***,
      "'listen' is deprecated. Please use the async 'start' or 'startCallback' method.",
      "DEP_WEBPACK_DEV_SERVER_LISTEN"
    )();

    if (typeof port === "function") {
      fn = port;
    ***REMOVED***

    if (
      typeof port !== "undefined" &&
      typeof this.options.port !== "undefined" &&
      port !== this.options.port
    ) {
      this.options.port = port;

      this.logger.warn(
        'The "port" specified in options is different from the port passed as an argument. Will be used from arguments.'
      );
    ***REMOVED***

    if (!this.options.port) {
      this.options.port = port;
    ***REMOVED***

    if (
      typeof hostname !== "undefined" &&
      typeof this.options.host !== "undefined" &&
      hostname !== this.options.host
    ) {
      this.options.host = hostname;

      this.logger.warn(
        'The "host" specified in options is different from the host passed as an argument. Will be used from arguments.'
      );
    ***REMOVED***

    if (!this.options.host) {
      this.options.host = hostname;
    ***REMOVED***

    this.start()
      .then(() => {
        if (fn) {
          fn.call(this.server);
        ***REMOVED***
      ***REMOVED***)
      .catch((error) => {
        // Nothing
        if (fn) {
          fn.call(this.server, error);
        ***REMOVED***
      ***REMOVED***);
  ***REMOVED***

  /**
   * @param {(err?: Error) => void***REMOVED*** [callback]
   * @returns {void***REMOVED***
   */
  // TODO remove in the next major release
  close(callback) {
    util.deprecate(
      () => {***REMOVED***,
      "'close' is deprecated. Please use the async 'stop' or 'stopCallback' method.",
      "DEP_WEBPACK_DEV_SERVER_CLOSE"
    )();

    this.stop()
      .then(() => {
        if (callback) {
          callback();
        ***REMOVED***
      ***REMOVED***)
      .catch((error) => {
        if (callback) {
          callback(error);
        ***REMOVED***
      ***REMOVED***);
  ***REMOVED***
***REMOVED***

module.exports = Server;
