// react-scripts/scripts/utils/verifyTypeScriptSetup.js
// @remove-file-on-eject
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

const chalk = require('react-dev-utils/chalk');
const fs = require('fs');
const resolve = require('resolve');
const path = require('path');
const paths = require('../../config/paths');
const os = require('os');
const semver = require('semver');
const immer = require('react-dev-utils/immer').produce;
const globby = require('react-dev-utils/globby').sync;

const hasJsxRuntime = (() => {
  if (process.env.DISABLE_NEW_JSX_TRANSFORM === 'true') {
    return false;
  ***REMOVED***

  try {
    require.resolve('react/jsx-runtime', { paths: [paths.appPath] ***REMOVED***);
    return true;
  ***REMOVED*** catch (e) {
    return false;
  ***REMOVED***
***REMOVED***)();

function writeJson(fileName, object) {
  fs.writeFileSync(
    fileName,
    JSON.stringify(object, null, 2).replace(/\n/g, os.EOL) + os.EOL
  );
***REMOVED***

function verifyNoTypeScript() {
  const typescriptFiles = globby(
    ['**/*.(ts|tsx)', '!**/node_modules', '!**/*.d.ts'],
    { cwd: paths.appSrc ***REMOVED***
  );
  if (typescriptFiles.length > 0) {
    console.warn(
      chalk.yellow(
        `We detected TypeScript in your project (${chalk.bold(
          `src${path.sep***REMOVED***${typescriptFiles[0]***REMOVED***`
        )***REMOVED***) and created a ${chalk.bold('tsconfig.json')***REMOVED*** file for you.`
      )
    );
    console.warn();
    return false;
  ***REMOVED***
  return true;
***REMOVED***

function verifyTypeScriptSetup() {
  let firstTimeSetup = false;

  if (!fs.existsSync(paths.appTsConfig)) {
    if (verifyNoTypeScript()) {
      return;
    ***REMOVED***
    writeJson(paths.appTsConfig, {***REMOVED***);
    firstTimeSetup = true;
  ***REMOVED***

  const isYarn = fs.existsSync(paths.yarnLockFile);

  // Ensure typescript is installed
  let ts;
  try {
    // TODO: Remove this hack once `globalThis` issue is resolved
    // https://github.com/jsdom/jsdom/issues/2961
    const globalThisWasDefined = !!global.globalThis;

    ts = require(resolve.sync('typescript', {
      basedir: paths.appNodeModules,
    ***REMOVED***));

    if (!globalThisWasDefined && !!global.globalThis) {
      delete global.globalThis;
    ***REMOVED***
  ***REMOVED*** catch (_) {
    console.error(
      chalk.bold.red(
        `It looks like you're trying to use TypeScript but do not have ${chalk.bold(
          'typescript'
        )***REMOVED*** installed.`
      )
    );
    console.error(
      chalk.bold(
        'Please install',
        chalk.cyan.bold('typescript'),
        'by running',
        chalk.cyan.bold(
          isYarn ? 'yarn add typescript' : 'npm install typescript'
        ) + '.'
      )
    );
    console.error(
      chalk.bold(
        'If you are not trying to use TypeScript, please remove the ' +
          chalk.cyan('tsconfig.json') +
          ' file from your package root (and any TypeScript files).'
      )
    );
    console.error();
    process.exit(1);
  ***REMOVED***

  const compilerOptions = {
    // These are suggested values and will be set when not present in the
    // tsconfig.json
    // 'parsedValue' matches the output value from ts.parseJsonConfigFileContent()
    target: {
      parsedValue: ts.ScriptTarget.ES5,
      suggested: 'es5',
    ***REMOVED***,
    lib: { suggested: ['dom', 'dom.iterable', 'esnext'] ***REMOVED***,
    allowJs: { suggested: true ***REMOVED***,
    skipLibCheck: { suggested: true ***REMOVED***,
    esModuleInterop: { suggested: true ***REMOVED***,
    allowSyntheticDefaultImports: { suggested: true ***REMOVED***,
    strict: { suggested: true ***REMOVED***,
    forceConsistentCasingInFileNames: { suggested: true ***REMOVED***,
    noFallthroughCasesInSwitch: { suggested: true ***REMOVED***,

    // These values are required and cannot be changed by the user
    // Keep this in sync with the webpack config
    module: {
      parsedValue: ts.ModuleKind.ESNext,
      value: 'esnext',
      reason: 'for import() and import/export',
    ***REMOVED***,
    moduleResolution: {
      parsedValue: ts.ModuleResolutionKind.NodeJs,
      value: 'node',
      reason: 'to match webpack resolution',
    ***REMOVED***,
    resolveJsonModule: { value: true, reason: 'to match webpack loader' ***REMOVED***,
    isolatedModules: { value: true, reason: 'implementation limitation' ***REMOVED***,
    noEmit: { value: true ***REMOVED***,
    jsx: {
      parsedValue:
        hasJsxRuntime && semver.gte(ts.version, '4.1.0-beta')
          ? ts.JsxEmit.ReactJSX
          : ts.JsxEmit.React,
      value:
        hasJsxRuntime && semver.gte(ts.version, '4.1.0-beta')
          ? 'react-jsx'
          : 'react',
      reason: 'to support the new JSX transform in React 17',
    ***REMOVED***,
    paths: { value: undefined, reason: 'aliased imports are not supported' ***REMOVED***,
  ***REMOVED***;

  const formatDiagnosticHost = {
    getCanonicalFileName: fileName => fileName,
    getCurrentDirectory: ts.sys.getCurrentDirectory,
    getNewLine: () => os.EOL,
  ***REMOVED***;

  const messages = [];
  let appTsConfig;
  let parsedTsConfig;
  let parsedCompilerOptions;
  try {
    const { config: readTsConfig, error ***REMOVED*** = ts.readConfigFile(
      paths.appTsConfig,
      ts.sys.readFile
    );

    if (error) {
      throw new Error(ts.formatDiagnostic(error, formatDiagnosticHost));
    ***REMOVED***

    appTsConfig = readTsConfig;

    // Get TS to parse and resolve any "extends"
    // Calling this function also mutates the tsconfig above,
    // adding in "include" and "exclude", but the compilerOptions remain untouched
    let result;
    parsedTsConfig = immer(readTsConfig, config => {
      result = ts.parseJsonConfigFileContent(
        config,
        ts.sys,
        path.dirname(paths.appTsConfig)
      );
    ***REMOVED***);

    if (result.errors && result.errors.length) {
      throw new Error(
        ts.formatDiagnostic(result.errors[0], formatDiagnosticHost)
      );
    ***REMOVED***

    parsedCompilerOptions = result.options;
  ***REMOVED*** catch (e) {
    if (e && e.name === 'SyntaxError') {
      console.error(
        chalk.red.bold(
          'Could not parse',
          chalk.cyan('tsconfig.json') + '.',
          'Please make sure it contains syntactically correct JSON.'
        )
      );
    ***REMOVED***

    console.log(e && e.message ? `${e.message***REMOVED***` : '');
    process.exit(1);
  ***REMOVED***

  if (appTsConfig.compilerOptions == null) {
    appTsConfig.compilerOptions = {***REMOVED***;
    firstTimeSetup = true;
  ***REMOVED***

  for (const option of Object.keys(compilerOptions)) {
    const { parsedValue, value, suggested, reason ***REMOVED*** = compilerOptions[option];

    const valueToCheck = parsedValue === undefined ? value : parsedValue;
    const coloredOption = chalk.cyan('compilerOptions.' + option);

    if (suggested != null) {
      if (parsedCompilerOptions[option] === undefined) {
        appTsConfig = immer(appTsConfig, config => {
          config.compilerOptions[option] = suggested;
        ***REMOVED***);
        messages.push(
          `${coloredOption***REMOVED*** to be ${chalk.bold(
            'suggested'
          )***REMOVED*** value: ${chalk.cyan.bold(suggested)***REMOVED*** (this can be changed)`
        );
      ***REMOVED***
    ***REMOVED*** else if (parsedCompilerOptions[option] !== valueToCheck) {
      appTsConfig = immer(appTsConfig, config => {
        config.compilerOptions[option] = value;
      ***REMOVED***);
      messages.push(
        `${coloredOption***REMOVED*** ${chalk.bold(
          valueToCheck == null ? 'must not' : 'must'
        )***REMOVED*** be ${valueToCheck == null ? 'set' : chalk.cyan.bold(value)***REMOVED***` +
          (reason != null ? ` (${reason***REMOVED***)` : '')
      );
    ***REMOVED***
  ***REMOVED***

  // tsconfig will have the merged "include" and "exclude" by this point
  if (parsedTsConfig.include == null) {
    appTsConfig = immer(appTsConfig, config => {
      config.include = ['src'];
    ***REMOVED***);
    messages.push(
      `${chalk.cyan('include')***REMOVED*** should be ${chalk.cyan.bold('src')***REMOVED***`
    );
  ***REMOVED***

  if (messages.length > 0) {
    if (firstTimeSetup) {
      console.log(
        chalk.bold(
          'Your',
          chalk.cyan('tsconfig.json'),
          'has been populated with default values.'
        )
      );
      console.log();
    ***REMOVED*** else {
      console.warn(
        chalk.bold(
          'The following changes are being made to your',
          chalk.cyan('tsconfig.json'),
          'file:'
        )
      );
      messages.forEach(message => {
        console.warn('  - ' + message);
      ***REMOVED***);
      console.warn();
    ***REMOVED***
    writeJson(paths.appTsConfig, appTsConfig);
  ***REMOVED***

  // Reference `react-scripts` types
  if (!fs.existsSync(paths.appTypeDeclarations)) {
    fs.writeFileSync(
      paths.appTypeDeclarations,
      `/// <reference types="react-scripts" />${os.EOL***REMOVED***`
    );
  ***REMOVED***
***REMOVED***

module.exports = verifyTypeScriptSetup;
