// webpack/lib/sharing/SharePlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra and Zackary Jackson @ScriptedAlchemy
*/

"use strict";

const { parseOptions ***REMOVED*** = require("../container/options");
const ConsumeSharedPlugin = require("./ConsumeSharedPlugin");
const ProvideSharedPlugin = require("./ProvideSharedPlugin");
const { isRequiredVersion ***REMOVED*** = require("./utils");

/** @typedef {import("../../declarations/plugins/sharing/ConsumeSharedPlugin").ConsumeSharedPluginOptions***REMOVED*** ConsumeSharedPluginOptions */
/** @typedef {import("../../declarations/plugins/sharing/ConsumeSharedPlugin").ConsumesConfig***REMOVED*** ConsumesConfig */
/** @typedef {import("../../declarations/plugins/sharing/ProvideSharedPlugin").ProvideSharedPluginOptions***REMOVED*** ProvideSharedPluginOptions */
/** @typedef {import("../../declarations/plugins/sharing/ProvideSharedPlugin").ProvidesConfig***REMOVED*** ProvidesConfig */
/** @typedef {import("../../declarations/plugins/sharing/SharePlugin").SharePluginOptions***REMOVED*** SharePluginOptions */
/** @typedef {import("../../declarations/plugins/sharing/SharePlugin").SharedConfig***REMOVED*** SharedConfig */
/** @typedef {import("../Compiler")***REMOVED*** Compiler */

class SharePlugin {
	/**
	 * @param {SharePluginOptions***REMOVED*** options options
	 */
	constructor(options) {
		/** @type {[string, SharedConfig][]***REMOVED*** */
		const sharedOptions = parseOptions(
			options.shared,
			(item, key) => {
				if (typeof item !== "string")
					throw new Error("Unexpected array in shared");
				/** @type {SharedConfig***REMOVED*** */
				const config =
					item === key || !isRequiredVersion(item)
						? {
								import: item
							***REMOVED***
						: {
								import: key,
								requiredVersion: item
							***REMOVED***;
				return config;
			***REMOVED***,
			item => item
		);
		/** @type {Record<string, ConsumesConfig>[]***REMOVED*** */
		const consumes = sharedOptions.map(([key, options]) => ({
			[key]: {
				import: options.import,
				shareKey: options.shareKey || key,
				shareScope: options.shareScope,
				requiredVersion: options.requiredVersion,
				strictVersion: options.strictVersion,
				singleton: options.singleton,
				packageName: options.packageName,
				eager: options.eager
			***REMOVED***
		***REMOVED***));
		/** @type {Record<string, ProvidesConfig>[]***REMOVED*** */
		const provides = sharedOptions
			.filter(([, options]) => options.import !== false)
			.map(([key, options]) => ({
				[options.import || key]: {
					shareKey: options.shareKey || key,
					shareScope: options.shareScope,
					version: options.version,
					eager: options.eager
				***REMOVED***
			***REMOVED***));
		this._shareScope = options.shareScope;
		this._consumes = consumes;
		this._provides = provides;
	***REMOVED***

	/**
	 * Apply the plugin
	 * @param {Compiler***REMOVED*** compiler the compiler instance
	 * @returns {void***REMOVED***
	 */
	apply(compiler) {
		new ConsumeSharedPlugin({
			shareScope: this._shareScope,
			consumes: this._consumes
		***REMOVED***).apply(compiler);
		new ProvideSharedPlugin({
			shareScope: this._shareScope,
			provides: this._provides
		***REMOVED***).apply(compiler);
	***REMOVED***
***REMOVED***

module.exports = SharePlugin;
