// postcss/lib/no-work-result.js
'use strict'

let MapGenerator = require('./map-generator')
let parse = require('./parse')
const Result = require('./result')
let stringify = require('./stringify')
let warnOnce = require('./warn-once')

class NoWorkResult {
  get content() {
    return this.result.css
  ***REMOVED***

  get css() {
    return this.result.css
  ***REMOVED***

  get map() {
    return this.result.map
  ***REMOVED***

  get messages() {
    return []
  ***REMOVED***

  get opts() {
    return this.result.opts
  ***REMOVED***

  get processor() {
    return this.result.processor
  ***REMOVED***

  get root() {
    if (this._root) {
      return this._root
    ***REMOVED***

    let root
    let parser = parse

    try {
      root = parser(this._css, this._opts)
    ***REMOVED*** catch (error) {
      this.error = error
    ***REMOVED***

    if (this.error) {
      throw this.error
    ***REMOVED*** else {
      this._root = root
      return root
    ***REMOVED***
  ***REMOVED***

  get [Symbol.toStringTag]() {
    return 'NoWorkResult'
  ***REMOVED***

  constructor(processor, css, opts) {
    css = css.toString()
    this.stringified = false

    this._processor = processor
    this._css = css
    this._opts = opts
    this._map = undefined
    let root

    let str = stringify
    this.result = new Result(this._processor, root, this._opts)
    this.result.css = css

    let self = this
    Object.defineProperty(this.result, 'root', {
      get() {
        return self.root
      ***REMOVED***
    ***REMOVED***)

    let map = new MapGenerator(str, root, this._opts, css)
    if (map.isMap()) {
      let [generatedCSS, generatedMap] = map.generate()
      if (generatedCSS) {
        this.result.css = generatedCSS
      ***REMOVED***
      if (generatedMap) {
        this.result.map = generatedMap
      ***REMOVED***
    ***REMOVED*** else {
      map.clearAnnotation()
      this.result.css = map.css
    ***REMOVED***
  ***REMOVED***

  async() {
    if (this.error) return Promise.reject(this.error)
    return Promise.resolve(this.result)
  ***REMOVED***

  catch(onRejected) {
    return this.async().catch(onRejected)
  ***REMOVED***

  finally(onFinally) {
    return this.async().then(onFinally, onFinally)
  ***REMOVED***

  sync() {
    if (this.error) throw this.error
    return this.result
  ***REMOVED***

  then(onFulfilled, onRejected) {
    if (process.env.NODE_ENV !== 'production') {
      if (!('from' in this._opts)) {
        warnOnce(
          'Without `from` option PostCSS could generate wrong source map ' +
            'and will not find Browserslist config. Set it to CSS file path ' +
            'or to `undefined` to prevent this warning.'
        )
      ***REMOVED***
    ***REMOVED***

    return this.async().then(onFulfilled, onRejected)
  ***REMOVED***

  toString() {
    return this._css
  ***REMOVED***

  warnings() {
    return []
  ***REMOVED***
***REMOVED***

module.exports = NoWorkResult
NoWorkResult.default = NoWorkResult
