// postcss/lib/root.js
'use strict'

let Container = require('./container')

let LazyResult, Processor

class Root extends Container {
  constructor(defaults) {
    super(defaults)
    this.type = 'root'
    if (!this.nodes) this.nodes = []
  ***REMOVED***

  normalize(child, sample, type) {
    let nodes = super.normalize(child)

    if (sample) {
      if (type === 'prepend') {
        if (this.nodes.length > 1) {
          sample.raws.before = this.nodes[1].raws.before
        ***REMOVED*** else {
          delete sample.raws.before
        ***REMOVED***
      ***REMOVED*** else if (this.first !== sample) {
        for (let node of nodes) {
          node.raws.before = sample.raws.before
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    return nodes
  ***REMOVED***

  removeChild(child, ignore) {
    let index = this.index(child)

    if (!ignore && index === 0 && this.nodes.length > 1) {
      this.nodes[1].raws.before = this.nodes[index].raws.before
    ***REMOVED***

    return super.removeChild(child)
  ***REMOVED***

  toResult(opts = {***REMOVED***) {
    let lazy = new LazyResult(new Processor(), this, opts)
    return lazy.stringify()
  ***REMOVED***
***REMOVED***

Root.registerLazyResult = dependant => {
  LazyResult = dependant
***REMOVED***

Root.registerProcessor = dependant => {
  Processor = dependant
***REMOVED***

module.exports = Root
Root.default = Root

Container.registerRoot(Root)
