// yargs/build/lib/usage.js
import { assertNotStrictEqual, ***REMOVED*** from './typings/common-types.js';
import { objFilter ***REMOVED*** from './utils/obj-filter.js';
import { YError ***REMOVED*** from './yerror.js';
import setBlocking from './utils/set-blocking.js';
export function usage(yargs, y18n, shim) {
    const __ = y18n.__;
    const self = {***REMOVED***;
    const fails = [];
    self.failFn = function failFn(f) {
        fails.push(f);
    ***REMOVED***;
    let failMessage = null;
    let showHelpOnFail = true;
    self.showHelpOnFail = function showHelpOnFailFn(arg1 = true, arg2) {
        function parseFunctionArgs() {
            return typeof arg1 === 'string' ? [true, arg1] : [arg1, arg2];
        ***REMOVED***
        const [enabled, message] = parseFunctionArgs();
        failMessage = message;
        showHelpOnFail = enabled;
        return self;
    ***REMOVED***;
    let failureOutput = false;
    self.fail = function fail(msg, err) {
        const logger = yargs._getLoggerInstance();
        if (fails.length) {
            for (let i = fails.length - 1; i >= 0; --i) {
                fails[i](msg, err, self);
            ***REMOVED***
        ***REMOVED***
        else {
            if (yargs.getExitProcess())
                setBlocking(true);
            if (!failureOutput) {
                failureOutput = true;
                if (showHelpOnFail) {
                    yargs.showHelp('error');
                    logger.error();
                ***REMOVED***
                if (msg || err)
                    logger.error(msg || err);
                if (failMessage) {
                    if (msg || err)
                        logger.error('');
                    logger.error(failMessage);
                ***REMOVED***
            ***REMOVED***
            err = err || new YError(msg);
            if (yargs.getExitProcess()) {
                return yargs.exit(1);
            ***REMOVED***
            else if (yargs._hasParseCallback()) {
                return yargs.exit(1, err);
            ***REMOVED***
            else {
                throw err;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    let usages = [];
    let usageDisabled = false;
    self.usage = (msg, description) => {
        if (msg === null) {
            usageDisabled = true;
            usages = [];
            return self;
        ***REMOVED***
        usageDisabled = false;
        usages.push([msg, description || '']);
        return self;
    ***REMOVED***;
    self.getUsage = () => {
        return usages;
    ***REMOVED***;
    self.getUsageDisabled = () => {
        return usageDisabled;
    ***REMOVED***;
    self.getPositionalGroupName = () => {
        return __('Positionals:');
    ***REMOVED***;
    let examples = [];
    self.example = (cmd, description) => {
        examples.push([cmd, description || '']);
    ***REMOVED***;
    let commands = [];
    self.command = function command(cmd, description, isDefault, aliases, deprecated = false) {
        if (isDefault) {
            commands = commands.map(cmdArray => {
                cmdArray[2] = false;
                return cmdArray;
            ***REMOVED***);
        ***REMOVED***
        commands.push([cmd, description || '', isDefault, aliases, deprecated]);
    ***REMOVED***;
    self.getCommands = () => commands;
    let descriptions = {***REMOVED***;
    self.describe = function describe(keyOrKeys, desc) {
        if (Array.isArray(keyOrKeys)) {
            keyOrKeys.forEach(k => {
                self.describe(k, desc);
            ***REMOVED***);
        ***REMOVED***
        else if (typeof keyOrKeys === 'object') {
            Object.keys(keyOrKeys).forEach(k => {
                self.describe(k, keyOrKeys[k]);
            ***REMOVED***);
        ***REMOVED***
        else {
            descriptions[keyOrKeys] = desc;
        ***REMOVED***
    ***REMOVED***;
    self.getDescriptions = () => descriptions;
    let epilogs = [];
    self.epilog = msg => {
        epilogs.push(msg);
    ***REMOVED***;
    let wrapSet = false;
    let wrap;
    self.wrap = cols => {
        wrapSet = true;
        wrap = cols;
    ***REMOVED***;
    function getWrap() {
        if (!wrapSet) {
            wrap = windowWidth();
            wrapSet = true;
        ***REMOVED***
        return wrap;
    ***REMOVED***
    const deferY18nLookupPrefix = '__yargsString__:';
    self.deferY18nLookup = str => deferY18nLookupPrefix + str;
    self.help = function help() {
        if (cachedHelpMessage)
            return cachedHelpMessage;
        normalizeAliases();
        const base$0 = yargs.customScriptName
            ? yargs.$0
            : shim.path.basename(yargs.$0);
        const demandedOptions = yargs.getDemandedOptions();
        const demandedCommands = yargs.getDemandedCommands();
        const deprecatedOptions = yargs.getDeprecatedOptions();
        const groups = yargs.getGroups();
        const options = yargs.getOptions();
        let keys = [];
        keys = keys.concat(Object.keys(descriptions));
        keys = keys.concat(Object.keys(demandedOptions));
        keys = keys.concat(Object.keys(demandedCommands));
        keys = keys.concat(Object.keys(options.default));
        keys = keys.filter(filterHiddenOptions);
        keys = Object.keys(keys.reduce((acc, key) => {
            if (key !== '_')
                acc[key] = true;
            return acc;
        ***REMOVED***, {***REMOVED***));
        const theWrap = getWrap();
        const ui = shim.cliui({
            width: theWrap,
            wrap: !!theWrap,
        ***REMOVED***);
        if (!usageDisabled) {
            if (usages.length) {
                usages.forEach(usage => {
                    ui.div(`${usage[0].replace(/\$0/g, base$0)***REMOVED***`);
                    if (usage[1]) {
                        ui.div({ text: `${usage[1]***REMOVED***`, padding: [1, 0, 0, 0] ***REMOVED***);
                    ***REMOVED***
                ***REMOVED***);
                ui.div();
            ***REMOVED***
            else if (commands.length) {
                let u = null;
                if (demandedCommands._) {
                    u = `${base$0***REMOVED*** <${__('command')***REMOVED***>\n`;
                ***REMOVED***
                else {
                    u = `${base$0***REMOVED*** [${__('command')***REMOVED***]\n`;
                ***REMOVED***
                ui.div(`${u***REMOVED***`);
            ***REMOVED***
        ***REMOVED***
        if (commands.length) {
            ui.div(__('Commands:'));
            const context = yargs.getContext();
            const parentCommands = context.commands.length
                ? `${context.commands.join(' ')***REMOVED*** `
                : '';
            if (yargs.getParserConfiguration()['sort-commands'] === true) {
                commands = commands.sort((a, b) => a[0].localeCompare(b[0]));
            ***REMOVED***
            commands.forEach(command => {
                const commandString = `${base$0***REMOVED*** ${parentCommands***REMOVED***${command[0].replace(/^\$0 ?/, '')***REMOVED***`;
                ui.span({
                    text: commandString,
                    padding: [0, 2, 0, 2],
                    width: maxWidth(commands, theWrap, `${base$0***REMOVED***${parentCommands***REMOVED***`) + 4,
                ***REMOVED***, { text: command[1] ***REMOVED***);
                const hints = [];
                if (command[2])
                    hints.push(`[${__('default')***REMOVED***]`);
                if (command[3] && command[3].length) {
                    hints.push(`[${__('aliases:')***REMOVED*** ${command[3].join(', ')***REMOVED***]`);
                ***REMOVED***
                if (command[4]) {
                    if (typeof command[4] === 'string') {
                        hints.push(`[${__('deprecated: %s', command[4])***REMOVED***]`);
                    ***REMOVED***
                    else {
                        hints.push(`[${__('deprecated')***REMOVED***]`);
                    ***REMOVED***
                ***REMOVED***
                if (hints.length) {
                    ui.div({
                        text: hints.join(' '),
                        padding: [0, 0, 0, 2],
                        align: 'right',
                    ***REMOVED***);
                ***REMOVED***
                else {
                    ui.div();
                ***REMOVED***
            ***REMOVED***);
            ui.div();
        ***REMOVED***
        const aliasKeys = (Object.keys(options.alias) || []).concat(Object.keys(yargs.parsed.newAliases) || []);
        keys = keys.filter(key => !yargs.parsed.newAliases[key] &&
            aliasKeys.every(alias => (options.alias[alias] || []).indexOf(key) === -1));
        const defaultGroup = __('Options:');
        if (!groups[defaultGroup])
            groups[defaultGroup] = [];
        addUngroupedKeys(keys, options.alias, groups, defaultGroup);
        const isLongSwitch = (sw) => /^--/.test(getText(sw));
        const displayedGroups = Object.keys(groups)
            .filter(groupName => groups[groupName].length > 0)
            .map(groupName => {
            const normalizedKeys = groups[groupName]
                .filter(filterHiddenOptions)
                .map(key => {
                if (~aliasKeys.indexOf(key))
                    return key;
                for (let i = 0, aliasKey; (aliasKey = aliasKeys[i]) !== undefined; i++) {
                    if (~(options.alias[aliasKey] || []).indexOf(key))
                        return aliasKey;
                ***REMOVED***
                return key;
            ***REMOVED***);
            return { groupName, normalizedKeys ***REMOVED***;
        ***REMOVED***)
            .filter(({ normalizedKeys ***REMOVED***) => normalizedKeys.length > 0)
            .map(({ groupName, normalizedKeys ***REMOVED***) => {
            const switches = normalizedKeys.reduce((acc, key) => {
                acc[key] = [key]
                    .concat(options.alias[key] || [])
                    .map(sw => {
                    if (groupName === self.getPositionalGroupName())
                        return sw;
                    else {
                        return ((/^[0-9]$/.test(sw)
                            ? ~options.boolean.indexOf(key)
                                ? '-'
                                : '--'
                            : sw.length > 1
                                ? '--'
                                : '-') + sw);
                    ***REMOVED***
                ***REMOVED***)
                    .sort((sw1, sw2) => isLongSwitch(sw1) === isLongSwitch(sw2)
                    ? 0
                    : isLongSwitch(sw1)
                        ? 1
                        : -1)
                    .join(', ');
                return acc;
            ***REMOVED***, {***REMOVED***);
            return { groupName, normalizedKeys, switches ***REMOVED***;
        ***REMOVED***);
        const shortSwitchesUsed = displayedGroups
            .filter(({ groupName ***REMOVED***) => groupName !== self.getPositionalGroupName())
            .some(({ normalizedKeys, switches ***REMOVED***) => !normalizedKeys.every(key => isLongSwitch(switches[key])));
        if (shortSwitchesUsed) {
            displayedGroups
                .filter(({ groupName ***REMOVED***) => groupName !== self.getPositionalGroupName())
                .forEach(({ normalizedKeys, switches ***REMOVED***) => {
                normalizedKeys.forEach(key => {
                    if (isLongSwitch(switches[key])) {
                        switches[key] = addIndentation(switches[key], '-x, '.length);
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***
        displayedGroups.forEach(({ groupName, normalizedKeys, switches ***REMOVED***) => {
            ui.div(groupName);
            normalizedKeys.forEach(key => {
                const kswitch = switches[key];
                let desc = descriptions[key] || '';
                let type = null;
                if (~desc.lastIndexOf(deferY18nLookupPrefix))
                    desc = __(desc.substring(deferY18nLookupPrefix.length));
                if (~options.boolean.indexOf(key))
                    type = `[${__('boolean')***REMOVED***]`;
                if (~options.count.indexOf(key))
                    type = `[${__('count')***REMOVED***]`;
                if (~options.string.indexOf(key))
                    type = `[${__('string')***REMOVED***]`;
                if (~options.normalize.indexOf(key))
                    type = `[${__('string')***REMOVED***]`;
                if (~options.array.indexOf(key))
                    type = `[${__('array')***REMOVED***]`;
                if (~options.number.indexOf(key))
                    type = `[${__('number')***REMOVED***]`;
                const deprecatedExtra = (deprecated) => typeof deprecated === 'string'
                    ? `[${__('deprecated: %s', deprecated)***REMOVED***]`
                    : `[${__('deprecated')***REMOVED***]`;
                const extra = [
                    key in deprecatedOptions
                        ? deprecatedExtra(deprecatedOptions[key])
                        : null,
                    type,
                    key in demandedOptions ? `[${__('required')***REMOVED***]` : null,
                    options.choices && options.choices[key]
                        ? `[${__('choices:')***REMOVED*** ${self.stringifiedValues(options.choices[key])***REMOVED***]`
                        : null,
                    defaultString(options.default[key], options.defaultDescription[key]),
                ]
                    .filter(Boolean)
                    .join(' ');
                ui.span({
                    text: getText(kswitch),
                    padding: [0, 2, 0, 2 + getIndentation(kswitch)],
                    width: maxWidth(switches, theWrap) + 4,
                ***REMOVED***, desc);
                if (extra)
                    ui.div({ text: extra, padding: [0, 0, 0, 2], align: 'right' ***REMOVED***);
                else
                    ui.div();
            ***REMOVED***);
            ui.div();
        ***REMOVED***);
        if (examples.length) {
            ui.div(__('Examples:'));
            examples.forEach(example => {
                example[0] = example[0].replace(/\$0/g, base$0);
            ***REMOVED***);
            examples.forEach(example => {
                if (example[1] === '') {
                    ui.div({
                        text: example[0],
                        padding: [0, 2, 0, 2],
                    ***REMOVED***);
                ***REMOVED***
                else {
                    ui.div({
                        text: example[0],
                        padding: [0, 2, 0, 2],
                        width: maxWidth(examples, theWrap) + 4,
                    ***REMOVED***, {
                        text: example[1],
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***);
            ui.div();
        ***REMOVED***
        if (epilogs.length > 0) {
            const e = epilogs
                .map(epilog => epilog.replace(/\$0/g, base$0))
                .join('\n');
            ui.div(`${e***REMOVED***\n`);
        ***REMOVED***
        return ui.toString().replace(/\s*$/, '');
    ***REMOVED***;
    function maxWidth(table, theWrap, modifier) {
        let width = 0;
        if (!Array.isArray(table)) {
            table = Object.values(table).map(v => [v]);
        ***REMOVED***
        table.forEach(v => {
            width = Math.max(shim.stringWidth(modifier ? `${modifier***REMOVED*** ${getText(v[0])***REMOVED***` : getText(v[0])) + getIndentation(v[0]), width);
        ***REMOVED***);
        if (theWrap)
            width = Math.min(width, parseInt((theWrap * 0.5).toString(), 10));
        return width;
    ***REMOVED***
    function normalizeAliases() {
        const demandedOptions = yargs.getDemandedOptions();
        const options = yargs.getOptions();
        (Object.keys(options.alias) || []).forEach(key => {
            options.alias[key].forEach(alias => {
                if (descriptions[alias])
                    self.describe(key, descriptions[alias]);
                if (alias in demandedOptions)
                    yargs.demandOption(key, demandedOptions[alias]);
                if (~options.boolean.indexOf(alias))
                    yargs.boolean(key);
                if (~options.count.indexOf(alias))
                    yargs.count(key);
                if (~options.string.indexOf(alias))
                    yargs.string(key);
                if (~options.normalize.indexOf(alias))
                    yargs.normalize(key);
                if (~options.array.indexOf(alias))
                    yargs.array(key);
                if (~options.number.indexOf(alias))
                    yargs.number(key);
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***
    let cachedHelpMessage;
    self.cacheHelpMessage = function () {
        cachedHelpMessage = this.help();
    ***REMOVED***;
    self.clearCachedHelpMessage = function () {
        cachedHelpMessage = undefined;
    ***REMOVED***;
    function addUngroupedKeys(keys, aliases, groups, defaultGroup) {
        let groupedKeys = [];
        let toCheck = null;
        Object.keys(groups).forEach(group => {
            groupedKeys = groupedKeys.concat(groups[group]);
        ***REMOVED***);
        keys.forEach(key => {
            toCheck = [key].concat(aliases[key]);
            if (!toCheck.some(k => groupedKeys.indexOf(k) !== -1)) {
                groups[defaultGroup].push(key);
            ***REMOVED***
        ***REMOVED***);
        return groupedKeys;
    ***REMOVED***
    function filterHiddenOptions(key) {
        return (yargs.getOptions().hiddenOptions.indexOf(key) < 0 ||
            yargs.parsed.argv[yargs.getOptions().showHiddenOpt]);
    ***REMOVED***
    self.showHelp = (level) => {
        const logger = yargs._getLoggerInstance();
        if (!level)
            level = 'error';
        const emit = typeof level === 'function' ? level : logger[level];
        emit(self.help());
    ***REMOVED***;
    self.functionDescription = fn => {
        const description = fn.name
            ? shim.Parser.decamelize(fn.name, '-')
            : __('generated-value');
        return ['(', description, ')'].join('');
    ***REMOVED***;
    self.stringifiedValues = function stringifiedValues(values, separator) {
        let string = '';
        const sep = separator || ', ';
        const array = [].concat(values);
        if (!values || !array.length)
            return string;
        array.forEach(value => {
            if (string.length)
                string += sep;
            string += JSON.stringify(value);
        ***REMOVED***);
        return string;
    ***REMOVED***;
    function defaultString(value, defaultDescription) {
        let string = `[${__('default:')***REMOVED*** `;
        if (value === undefined && !defaultDescription)
            return null;
        if (defaultDescription) {
            string += defaultDescription;
        ***REMOVED***
        else {
            switch (typeof value) {
                case 'string':
                    string += `"${value***REMOVED***"`;
                    break;
                case 'object':
                    string += JSON.stringify(value);
                    break;
                default:
                    string += value;
            ***REMOVED***
        ***REMOVED***
        return `${string***REMOVED***]`;
    ***REMOVED***
    function windowWidth() {
        const maxWidth = 80;
        if (shim.process.stdColumns) {
            return Math.min(maxWidth, shim.process.stdColumns);
        ***REMOVED***
        else {
            return maxWidth;
        ***REMOVED***
    ***REMOVED***
    let version = null;
    self.version = ver => {
        version = ver;
    ***REMOVED***;
    self.showVersion = () => {
        const logger = yargs._getLoggerInstance();
        logger.log(version);
    ***REMOVED***;
    self.reset = function reset(localLookup) {
        failMessage = null;
        failureOutput = false;
        usages = [];
        usageDisabled = false;
        epilogs = [];
        examples = [];
        commands = [];
        descriptions = objFilter(descriptions, k => !localLookup[k]);
        return self;
    ***REMOVED***;
    const frozens = [];
    self.freeze = function freeze() {
        frozens.push({
            failMessage,
            failureOutput,
            usages,
            usageDisabled,
            epilogs,
            examples,
            commands,
            descriptions,
        ***REMOVED***);
    ***REMOVED***;
    self.unfreeze = function unfreeze() {
        const frozen = frozens.pop();
        assertNotStrictEqual(frozen, undefined, shim);
        ({
            failMessage,
            failureOutput,
            usages,
            usageDisabled,
            epilogs,
            examples,
            commands,
            descriptions,
        ***REMOVED*** = frozen);
    ***REMOVED***;
    return self;
***REMOVED***
function isIndentedText(text) {
    return typeof text === 'object';
***REMOVED***
function addIndentation(text, indent) {
    return isIndentedText(text)
        ? { text: text.text, indentation: text.indentation + indent ***REMOVED***
        : { text, indentation: indent ***REMOVED***;
***REMOVED***
function getIndentation(text) {
    return isIndentedText(text) ? text.indentation : 0;
***REMOVED***
function getText(text) {
    return isIndentedText(text) ? text.text : text;
***REMOVED***
