// yargs/build/lib/validation.js
import { argsert ***REMOVED*** from './argsert.js';
import { assertNotStrictEqual, ***REMOVED*** from './typings/common-types.js';
import { levenshtein as distance ***REMOVED*** from './utils/levenshtein.js';
import { objFilter ***REMOVED*** from './utils/obj-filter.js';
const specialKeys = ['$0', '--', '_'];
export function validation(yargs, usage, y18n, shim) {
    const __ = y18n.__;
    const __n = y18n.__n;
    const self = {***REMOVED***;
    self.nonOptionCount = function nonOptionCount(argv) {
        const demandedCommands = yargs.getDemandedCommands();
        const positionalCount = argv._.length + (argv['--'] ? argv['--'].length : 0);
        const _s = positionalCount - yargs.getContext().commands.length;
        if (demandedCommands._ &&
            (_s < demandedCommands._.min || _s > demandedCommands._.max)) {
            if (_s < demandedCommands._.min) {
                if (demandedCommands._.minMsg !== undefined) {
                    usage.fail(demandedCommands._.minMsg
                        ? demandedCommands._.minMsg
                            .replace(/\$0/g, _s.toString())
                            .replace(/\$1/, demandedCommands._.min.toString())
                        : null);
                ***REMOVED***
                else {
                    usage.fail(__n('Not enough non-option arguments: got %s, need at least %s', 'Not enough non-option arguments: got %s, need at least %s', _s, _s.toString(), demandedCommands._.min.toString()));
                ***REMOVED***
            ***REMOVED***
            else if (_s > demandedCommands._.max) {
                if (demandedCommands._.maxMsg !== undefined) {
                    usage.fail(demandedCommands._.maxMsg
                        ? demandedCommands._.maxMsg
                            .replace(/\$0/g, _s.toString())
                            .replace(/\$1/, demandedCommands._.max.toString())
                        : null);
                ***REMOVED***
                else {
                    usage.fail(__n('Too many non-option arguments: got %s, maximum of %s', 'Too many non-option arguments: got %s, maximum of %s', _s, _s.toString(), demandedCommands._.max.toString()));
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    self.positionalCount = function positionalCount(required, observed) {
        if (observed < required) {
            usage.fail(__n('Not enough non-option arguments: got %s, need at least %s', 'Not enough non-option arguments: got %s, need at least %s', observed, observed + '', required + ''));
        ***REMOVED***
    ***REMOVED***;
    self.requiredArguments = function requiredArguments(argv) {
        const demandedOptions = yargs.getDemandedOptions();
        let missing = null;
        for (const key of Object.keys(demandedOptions)) {
            if (!Object.prototype.hasOwnProperty.call(argv, key) ||
                typeof argv[key] === 'undefined') {
                missing = missing || {***REMOVED***;
                missing[key] = demandedOptions[key];
            ***REMOVED***
        ***REMOVED***
        if (missing) {
            const customMsgs = [];
            for (const key of Object.keys(missing)) {
                const msg = missing[key];
                if (msg && customMsgs.indexOf(msg) < 0) {
                    customMsgs.push(msg);
                ***REMOVED***
            ***REMOVED***
            const customMsg = customMsgs.length ? `\n${customMsgs.join('\n')***REMOVED***` : '';
            usage.fail(__n('Missing required argument: %s', 'Missing required arguments: %s', Object.keys(missing).length, Object.keys(missing).join(', ') + customMsg));
        ***REMOVED***
    ***REMOVED***;
    self.unknownArguments = function unknownArguments(argv, aliases, positionalMap, isDefaultCommand, checkPositionals = true) {
        const commandKeys = yargs.getCommandInstance().getCommands();
        const unknown = [];
        const currentContext = yargs.getContext();
        Object.keys(argv).forEach(key => {
            if (specialKeys.indexOf(key) === -1 &&
                !Object.prototype.hasOwnProperty.call(positionalMap, key) &&
                !Object.prototype.hasOwnProperty.call(yargs._getParseContext(), key) &&
                !self.isValidAndSomeAliasIsNotNew(key, aliases)) {
                unknown.push(key);
            ***REMOVED***
        ***REMOVED***);
        if (checkPositionals &&
            (currentContext.commands.length > 0 ||
                commandKeys.length > 0 ||
                isDefaultCommand)) {
            argv._.slice(currentContext.commands.length).forEach(key => {
                if (commandKeys.indexOf('' + key) === -1) {
                    unknown.push('' + key);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***
        if (unknown.length > 0) {
            usage.fail(__n('Unknown argument: %s', 'Unknown arguments: %s', unknown.length, unknown.join(', ')));
        ***REMOVED***
    ***REMOVED***;
    self.unknownCommands = function unknownCommands(argv) {
        const commandKeys = yargs.getCommandInstance().getCommands();
        const unknown = [];
        const currentContext = yargs.getContext();
        if (currentContext.commands.length > 0 || commandKeys.length > 0) {
            argv._.slice(currentContext.commands.length).forEach(key => {
                if (commandKeys.indexOf('' + key) === -1) {
                    unknown.push('' + key);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***
        if (unknown.length > 0) {
            usage.fail(__n('Unknown command: %s', 'Unknown commands: %s', unknown.length, unknown.join(', ')));
            return true;
        ***REMOVED***
        else {
            return false;
        ***REMOVED***
    ***REMOVED***;
    self.isValidAndSomeAliasIsNotNew = function isValidAndSomeAliasIsNotNew(key, aliases) {
        if (!Object.prototype.hasOwnProperty.call(aliases, key)) {
            return false;
        ***REMOVED***
        const newAliases = yargs.parsed.newAliases;
        for (const a of [key, ...aliases[key]]) {
            if (!Object.prototype.hasOwnProperty.call(newAliases, a) ||
                !newAliases[key]) {
                return true;
            ***REMOVED***
        ***REMOVED***
        return false;
    ***REMOVED***;
    self.limitedChoices = function limitedChoices(argv) {
        const options = yargs.getOptions();
        const invalid = {***REMOVED***;
        if (!Object.keys(options.choices).length)
            return;
        Object.keys(argv).forEach(key => {
            if (specialKeys.indexOf(key) === -1 &&
                Object.prototype.hasOwnProperty.call(options.choices, key)) {
                [].concat(argv[key]).forEach(value => {
                    if (options.choices[key].indexOf(value) === -1 &&
                        value !== undefined) {
                        invalid[key] = (invalid[key] || []).concat(value);
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***);
        const invalidKeys = Object.keys(invalid);
        if (!invalidKeys.length)
            return;
        let msg = __('Invalid values:');
        invalidKeys.forEach(key => {
            msg += `\n  ${__('Argument: %s, Given: %s, Choices: %s', key, usage.stringifiedValues(invalid[key]), usage.stringifiedValues(options.choices[key]))***REMOVED***`;
        ***REMOVED***);
        usage.fail(msg);
    ***REMOVED***;
    let checks = [];
    self.check = function check(f, global) {
        checks.push({
            func: f,
            global,
        ***REMOVED***);
    ***REMOVED***;
    self.customChecks = function customChecks(argv, aliases) {
        for (let i = 0, f; (f = checks[i]) !== undefined; i++) {
            const func = f.func;
            let result = null;
            try {
                result = func(argv, aliases);
            ***REMOVED***
            catch (err) {
                usage.fail(err.message ? err.message : err, err);
                continue;
            ***REMOVED***
            if (!result) {
                usage.fail(__('Argument check failed: %s', func.toString()));
            ***REMOVED***
            else if (typeof result === 'string' || result instanceof Error) {
                usage.fail(result.toString(), result);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    let implied = {***REMOVED***;
    self.implies = function implies(key, value) {
        argsert('<string|object> [array|number|string]', [key, value], arguments.length);
        if (typeof key === 'object') {
            Object.keys(key).forEach(k => {
                self.implies(k, key[k]);
            ***REMOVED***);
        ***REMOVED***
        else {
            yargs.global(key);
            if (!implied[key]) {
                implied[key] = [];
            ***REMOVED***
            if (Array.isArray(value)) {
                value.forEach(i => self.implies(key, i));
            ***REMOVED***
            else {
                assertNotStrictEqual(value, undefined, shim);
                implied[key].push(value);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    self.getImplied = function getImplied() {
        return implied;
    ***REMOVED***;
    function keyExists(argv, val) {
        const num = Number(val);
        val = isNaN(num) ? val : num;
        if (typeof val === 'number') {
            val = argv._.length >= val;
        ***REMOVED***
        else if (val.match(/^--no-.+/)) {
            val = val.match(/^--no-(.+)/)[1];
            val = !argv[val];
        ***REMOVED***
        else {
            val = argv[val];
        ***REMOVED***
        return val;
    ***REMOVED***
    self.implications = function implications(argv) {
        const implyFail = [];
        Object.keys(implied).forEach(key => {
            const origKey = key;
            (implied[key] || []).forEach(value => {
                let key = origKey;
                const origValue = value;
                key = keyExists(argv, key);
                value = keyExists(argv, value);
                if (key && !value) {
                    implyFail.push(` ${origKey***REMOVED*** -> ${origValue***REMOVED***`);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
        if (implyFail.length) {
            let msg = `${__('Implications failed:')***REMOVED***\n`;
            implyFail.forEach(value => {
                msg += value;
            ***REMOVED***);
            usage.fail(msg);
        ***REMOVED***
    ***REMOVED***;
    let conflicting = {***REMOVED***;
    self.conflicts = function conflicts(key, value) {
        argsert('<string|object> [array|string]', [key, value], arguments.length);
        if (typeof key === 'object') {
            Object.keys(key).forEach(k => {
                self.conflicts(k, key[k]);
            ***REMOVED***);
        ***REMOVED***
        else {
            yargs.global(key);
            if (!conflicting[key]) {
                conflicting[key] = [];
            ***REMOVED***
            if (Array.isArray(value)) {
                value.forEach(i => self.conflicts(key, i));
            ***REMOVED***
            else {
                conflicting[key].push(value);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    self.getConflicting = () => conflicting;
    self.conflicting = function conflictingFn(argv) {
        Object.keys(argv).forEach(key => {
            if (conflicting[key]) {
                conflicting[key].forEach(value => {
                    if (value && argv[key] !== undefined && argv[value] !== undefined) {
                        usage.fail(__('Arguments %s and %s are mutually exclusive', key, value));
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***;
    self.recommendCommands = function recommendCommands(cmd, potentialCommands) {
        const threshold = 3;
        potentialCommands = potentialCommands.sort((a, b) => b.length - a.length);
        let recommended = null;
        let bestDistance = Infinity;
        for (let i = 0, candidate; (candidate = potentialCommands[i]) !== undefined; i++) {
            const d = distance(cmd, candidate);
            if (d <= threshold && d < bestDistance) {
                bestDistance = d;
                recommended = candidate;
            ***REMOVED***
        ***REMOVED***
        if (recommended)
            usage.fail(__('Did you mean %s?', recommended));
    ***REMOVED***;
    self.reset = function reset(localLookup) {
        implied = objFilter(implied, k => !localLookup[k]);
        conflicting = objFilter(conflicting, k => !localLookup[k]);
        checks = checks.filter(c => c.global);
        return self;
    ***REMOVED***;
    const frozens = [];
    self.freeze = function freeze() {
        frozens.push({
            implied,
            checks,
            conflicting,
        ***REMOVED***);
    ***REMOVED***;
    self.unfreeze = function unfreeze() {
        const frozen = frozens.pop();
        assertNotStrictEqual(frozen, undefined, shim);
        ({ implied, checks, conflicting ***REMOVED*** = frozen);
    ***REMOVED***;
    return self;
***REMOVED***
