// bluebird/js/release/reduce.js
"use strict";
module.exports = function(Promise,
                          PromiseArray,
                          apiRejection,
                          tryConvertToPromise,
                          INTERNAL,
                          debug) {
var util = require("./util");
var tryCatch = util.tryCatch;

function ReductionPromiseArray(promises, fn, initialValue, _each) {
    this.constructor$(promises);
    var context = Promise._getContext();
    this._fn = util.contextBind(context, fn);
    if (initialValue !== undefined) {
        initialValue = Promise.resolve(initialValue);
        initialValue._attachCancellationCallback(this);
    ***REMOVED***
    this._initialValue = initialValue;
    this._currentCancellable = null;
    if(_each === INTERNAL) {
        this._eachValues = Array(this._length);
    ***REMOVED*** else if (_each === 0) {
        this._eachValues = null;
    ***REMOVED*** else {
        this._eachValues = undefined;
    ***REMOVED***
    this._promise._captureStackTrace();
    this._init$(undefined, -5);
***REMOVED***
util.inherits(ReductionPromiseArray, PromiseArray);

ReductionPromiseArray.prototype._gotAccum = function(accum) {
    if (this._eachValues !== undefined &&
        this._eachValues !== null &&
        accum !== INTERNAL) {
        this._eachValues.push(accum);
    ***REMOVED***
***REMOVED***;

ReductionPromiseArray.prototype._eachComplete = function(value) {
    if (this._eachValues !== null) {
        this._eachValues.push(value);
    ***REMOVED***
    return this._eachValues;
***REMOVED***;

ReductionPromiseArray.prototype._init = function() {***REMOVED***;

ReductionPromiseArray.prototype._resolveEmptyArray = function() {
    this._resolve(this._eachValues !== undefined ? this._eachValues
                                                 : this._initialValue);
***REMOVED***;

ReductionPromiseArray.prototype.shouldCopyValues = function () {
    return false;
***REMOVED***;

ReductionPromiseArray.prototype._resolve = function(value) {
    this._promise._resolveCallback(value);
    this._values = null;
***REMOVED***;

ReductionPromiseArray.prototype._resultCancelled = function(sender) {
    if (sender === this._initialValue) return this._cancel();
    if (this._isResolved()) return;
    this._resultCancelled$();
    if (this._currentCancellable instanceof Promise) {
        this._currentCancellable.cancel();
    ***REMOVED***
    if (this._initialValue instanceof Promise) {
        this._initialValue.cancel();
    ***REMOVED***
***REMOVED***;

ReductionPromiseArray.prototype._iterate = function (values) {
    this._values = values;
    var value;
    var i;
    var length = values.length;
    if (this._initialValue !== undefined) {
        value = this._initialValue;
        i = 0;
    ***REMOVED*** else {
        value = Promise.resolve(values[0]);
        i = 1;
    ***REMOVED***

    this._currentCancellable = value;

    for (var j = i; j < length; ++j) {
        var maybePromise = values[j];
        if (maybePromise instanceof Promise) {
            maybePromise.suppressUnhandledRejections();
        ***REMOVED***
    ***REMOVED***

    if (!value.isRejected()) {
        for (; i < length; ++i) {
            var ctx = {
                accum: null,
                value: values[i],
                index: i,
                length: length,
                array: this
            ***REMOVED***;

            value = value._then(gotAccum, undefined, undefined, ctx, undefined);

            if ((i & 127) === 0) {
                value._setNoAsyncGuarantee();
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    if (this._eachValues !== undefined) {
        value = value
            ._then(this._eachComplete, undefined, undefined, this, undefined);
    ***REMOVED***
    value._then(completed, completed, undefined, value, this);
***REMOVED***;

Promise.prototype.reduce = function (fn, initialValue) {
    return reduce(this, fn, initialValue, null);
***REMOVED***;

Promise.reduce = function (promises, fn, initialValue, _each) {
    return reduce(promises, fn, initialValue, _each);
***REMOVED***;

function completed(valueOrReason, array) {
    if (this.isFulfilled()) {
        array._resolve(valueOrReason);
    ***REMOVED*** else {
        array._reject(valueOrReason);
    ***REMOVED***
***REMOVED***

function reduce(promises, fn, initialValue, _each) {
    if (typeof fn !== "function") {
        return apiRejection("expecting a function but got " + util.classString(fn));
    ***REMOVED***
    var array = new ReductionPromiseArray(promises, fn, initialValue, _each);
    return array.promise();
***REMOVED***

function gotAccum(accum) {
    this.accum = accum;
    this.array._gotAccum(accum);
    var value = tryConvertToPromise(this.value, this.array._promise);
    if (value instanceof Promise) {
        this.array._currentCancellable = value;
        return value._then(gotValue, undefined, undefined, this, undefined);
    ***REMOVED*** else {
        return gotValue.call(this, value);
    ***REMOVED***
***REMOVED***

function gotValue(value) {
    var array = this.array;
    var promise = array._promise;
    var fn = tryCatch(array._fn);
    promise._pushContext();
    var ret;
    if (array._eachValues !== undefined) {
        ret = fn.call(promise._boundValue(), value, this.index, this.length);
    ***REMOVED*** else {
        ret = fn.call(promise._boundValue(),
                              this.accum, value, this.index, this.length);
    ***REMOVED***
    if (ret instanceof Promise) {
        array._currentCancellable = ret;
    ***REMOVED***
    var promiseCreated = promise._popContext();
    debug.checkForgottenReturns(
        ret,
        promiseCreated,
        array._eachValues !== undefined ? "Promise.each" : "Promise.reduce",
        promise
    );
    return ret;
***REMOVED***
***REMOVED***;
