// jsdom/lib/jsdom/living/file-api/Blob-impl.js
"use strict";
const Blob = require("../generated/Blob");
const { isArrayBuffer ***REMOVED*** = require("../generated/utils");

function convertLineEndingsToNative(s) {
  // jsdom always pretends to be *nix, for consistency.
  // See also https://github.com/jsdom/jsdom/issues/2396.
  return s.replace(/\r\n|\r|\n/g, "\n");
***REMOVED***

exports.implementation = class BlobImpl {
  constructor(globalObject, args) {
    const parts = args[0];
    const properties = args[1];

    const buffers = [];

    if (parts !== undefined) {
      for (const part of parts) {
        let buffer;
        if (isArrayBuffer(part)) {
          buffer = Buffer.from(part);
        ***REMOVED*** else if (ArrayBuffer.isView(part)) {
          buffer = Buffer.from(part.buffer, part.byteOffset, part.byteLength);
        ***REMOVED*** else if (Blob.isImpl(part)) {
          buffer = part._buffer;
        ***REMOVED*** else {
          let s = part;
          if (properties.endings === "native") {
            s = convertLineEndingsToNative(part);
          ***REMOVED***
          buffer = Buffer.from(s);
        ***REMOVED***
        buffers.push(buffer);
      ***REMOVED***
    ***REMOVED***

    this._buffer = Buffer.concat(buffers);
    this._globalObject = globalObject;

    this.type = properties.type;
    if (/[^\u0020-\u007E]/.test(this.type)) {
      this.type = "";
    ***REMOVED*** else {
      this.type = this.type.toLowerCase();
    ***REMOVED***
  ***REMOVED***

  get size() {
    return this._buffer.length;
  ***REMOVED***

  slice(start, end, contentType) {
    const { size ***REMOVED*** = this;

    let relativeStart, relativeEnd, relativeContentType;

    if (start === undefined) {
      relativeStart = 0;
    ***REMOVED*** else if (start < 0) {
      relativeStart = Math.max(size + start, 0);
    ***REMOVED*** else {
      relativeStart = Math.min(start, size);
    ***REMOVED***
    if (end === undefined) {
      relativeEnd = size;
    ***REMOVED*** else if (end < 0) {
      relativeEnd = Math.max(size + end, 0);
    ***REMOVED*** else {
      relativeEnd = Math.min(end, size);
    ***REMOVED***

    if (contentType === undefined) {
      relativeContentType = "";
    ***REMOVED*** else {
      // sanitization (lower case and invalid char check) is done in the
      // constructor
      relativeContentType = contentType;
    ***REMOVED***

    const span = Math.max(relativeEnd - relativeStart, 0);

    const buffer = this._buffer;
    const slicedBuffer = buffer.slice(
      relativeStart,
      relativeStart + span
    );

    const blob = Blob.createImpl(this._globalObject, [[], { type: relativeContentType ***REMOVED***], {***REMOVED***);
    blob._buffer = slicedBuffer;
    return blob;
  ***REMOVED***
***REMOVED***;
