// jsdom/lib/jsdom/living/file-api/FileReader-impl.js
"use strict";

const whatwgEncoding = require("whatwg-encoding");
const MIMEType = require("whatwg-mimetype");
const DOMException = require("domexception/webidl2js-wrapper");
const EventTargetImpl = require("../events/EventTarget-impl").implementation;
const ProgressEvent = require("../generated/ProgressEvent");
const { setupForSimpleEventAccessors ***REMOVED*** = require("../helpers/create-event-accessor");
const { fireAnEvent ***REMOVED*** = require("../helpers/events");
const { copyToArrayBufferInNewRealm ***REMOVED*** = require("../helpers/binary-data");

const READY_STATES = Object.freeze({
  EMPTY: 0,
  LOADING: 1,
  DONE: 2
***REMOVED***);

const events = ["loadstart", "progress", "load", "abort", "error", "loadend"];

class FileReaderImpl extends EventTargetImpl {
  constructor(globalObject, args, privateData) {
    super(globalObject, args, privateData);

    this.error = null;
    this.readyState = READY_STATES.EMPTY;
    this.result = null;

    this._globalObject = globalObject;
    this._ownerDocument = globalObject.document;
    this._terminated = false;
  ***REMOVED***

  readAsArrayBuffer(file) {
    this._readFile(file, "buffer");
  ***REMOVED***
  readAsBinaryString(file) {
    this._readFile(file, "binaryString");
  ***REMOVED***
  readAsDataURL(file) {
    this._readFile(file, "dataURL");
  ***REMOVED***
  readAsText(file, encoding) {
    this._readFile(file, "text", whatwgEncoding.labelToName(encoding) || "UTF-8");
  ***REMOVED***

  abort() {
    if (this.readyState === READY_STATES.EMPTY || this.readyState === READY_STATES.DONE) {
      this.result = null;
      return;
    ***REMOVED***

    if (this.readyState === READY_STATES.LOADING) {
      this.readyState = READY_STATES.DONE;
      this.result = null;
    ***REMOVED***

    this._terminated = true;
    this._fireProgressEvent("abort");
    this._fireProgressEvent("loadend");
  ***REMOVED***

  _fireProgressEvent(name, props) {
    fireAnEvent(name, this, ProgressEvent, props);
  ***REMOVED***

  _readFile(file, format, encoding) {
    if (this.readyState === READY_STATES.LOADING) {
      throw DOMException.create(this._globalObject, [
        "The object is in an invalid state.",
        "InvalidStateError"
      ]);
    ***REMOVED***

    this.readyState = READY_STATES.LOADING;

    setImmediate(() => {
      if (this._terminated) {
        this._terminated = false;
        return;
      ***REMOVED***

      this._fireProgressEvent("loadstart");

      let data = file._buffer;
      if (!data) {
        data = Buffer.alloc(0);
      ***REMOVED***
      this._fireProgressEvent("progress", {
        lengthComputable: !isNaN(file.size),
        total: file.size,
        loaded: data.length
      ***REMOVED***);

      setImmediate(() => {
        if (this._terminated) {
          this._terminated = false;
          return;
        ***REMOVED***

        switch (format) {
          case "binaryString": {
            this.result = data.toString("binary");
            break;
          ***REMOVED***
          case "dataURL": {
            // Spec seems very unclear here; see https://github.com/w3c/FileAPI/issues/104.
            const contentType = MIMEType.parse(file.type) || "application/octet-stream";
            this.result = `data:${contentType***REMOVED***;base64,${data.toString("base64")***REMOVED***`;
            break;
          ***REMOVED***
          case "text": {
            this.result = whatwgEncoding.decode(data, encoding);
            break;
          ***REMOVED***
          case "buffer":
          default: {
            this.result = copyToArrayBufferInNewRealm(data, this._globalObject);
            break;
          ***REMOVED***
        ***REMOVED***
        this.readyState = READY_STATES.DONE;
        this._fireProgressEvent("load");
        this._fireProgressEvent("loadend");
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***
***REMOVED***
setupForSimpleEventAccessors(FileReaderImpl.prototype, events);

exports.implementation = FileReaderImpl;
