// ajv/lib/vocabularies/jtd/ref.ts
import type {CodeKeywordDefinition, AnySchemaObject***REMOVED*** from "../../types"
import type {KeywordCxt***REMOVED*** from "../../compile/validate"
import {compileSchema, SchemaEnv***REMOVED*** from "../../compile"
import {_, not, nil, stringify***REMOVED*** from "../../compile/codegen"
import MissingRefError from "../../compile/ref_error"
import N from "../../compile/names"
import {getValidate, callRef***REMOVED*** from "../core/ref"
import {checkMetadata***REMOVED*** from "./metadata"

const def: CodeKeywordDefinition = {
  keyword: "ref",
  schemaType: "string",
  code(cxt: KeywordCxt) {
    checkMetadata(cxt)
    const {gen, data, schema: ref, parentSchema, it***REMOVED*** = cxt
    const {
      schemaEnv: {root***REMOVED***,
    ***REMOVED*** = it
    const valid = gen.name("valid")
    if (parentSchema.nullable) {
      gen.var(valid, _`${data***REMOVED*** === null`)
      gen.if(not(valid), validateJtdRef)
    ***REMOVED*** else {
      gen.var(valid, false)
      validateJtdRef()
    ***REMOVED***
    cxt.ok(valid)

    function validateJtdRef(): void {
      const refSchema = (root.schema as AnySchemaObject).definitions?.[ref]
      if (!refSchema) {
        throw new MissingRefError(it.opts.uriResolver, "", ref, `No definition ${ref***REMOVED***`)
      ***REMOVED***
      if (hasRef(refSchema) || !it.opts.inlineRefs) callValidate(refSchema)
      else inlineRefSchema(refSchema)
    ***REMOVED***

    function callValidate(schema: AnySchemaObject): void {
      const sch = compileSchema.call(
        it.self,
        new SchemaEnv({schema, root, schemaPath: `/definitions/${ref***REMOVED***`***REMOVED***)
      )
      const v = getValidate(cxt, sch)
      const errsCount = gen.const("_errs", N.errors)
      callRef(cxt, v, sch, sch.$async)
      gen.assign(valid, _`${errsCount***REMOVED*** === ${N.errors***REMOVED***`)
    ***REMOVED***

    function inlineRefSchema(schema: AnySchemaObject): void {
      const schName = gen.scopeValue(
        "schema",
        it.opts.code.source === true ? {ref: schema, code: stringify(schema)***REMOVED*** : {ref: schema***REMOVED***
      )
      cxt.subschema(
        {
          schema,
          dataTypes: [],
          schemaPath: nil,
          topSchemaRef: schName,
          errSchemaPath: `/definitions/${ref***REMOVED***`,
        ***REMOVED***,
        valid
      )
    ***REMOVED***
  ***REMOVED***,
***REMOVED***

export function hasRef(schema: AnySchemaObject): boolean {
  for (const key in schema) {
    let sch: AnySchemaObject
    if (key === "ref" || (typeof (sch = schema[key]) == "object" && hasRef(sch))) return true
  ***REMOVED***
  return false
***REMOVED***

export default def
