// ajv/dist/2020.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.MissingRefError = exports.ValidationError = exports.CodeGen = exports.Name = exports.nil = exports.stringify = exports.str = exports._ = exports.KeywordCxt = exports.Ajv2020 = void 0;
const core_1 = require("./core");
const draft2020_1 = require("./vocabularies/draft2020");
const discriminator_1 = require("./vocabularies/discriminator");
const json_schema_2020_12_1 = require("./refs/json-schema-2020-12");
const META_SCHEMA_ID = "https://json-schema.org/draft/2020-12/schema";
class Ajv2020 extends core_1.default {
    constructor(opts = {***REMOVED***) {
        super({
            ...opts,
            dynamicRef: true,
            next: true,
            unevaluated: true,
        ***REMOVED***);
    ***REMOVED***
    _addVocabularies() {
        super._addVocabularies();
        draft2020_1.default.forEach((v) => this.addVocabulary(v));
        if (this.opts.discriminator)
            this.addKeyword(discriminator_1.default);
    ***REMOVED***
    _addDefaultMetaSchema() {
        super._addDefaultMetaSchema();
        const { $data, meta ***REMOVED*** = this.opts;
        if (!meta)
            return;
        json_schema_2020_12_1.default.call(this, $data);
        this.refs["http://json-schema.org/schema"] = META_SCHEMA_ID;
    ***REMOVED***
    defaultMeta() {
        return (this.opts.defaultMeta =
            super.defaultMeta() || (this.getSchema(META_SCHEMA_ID) ? META_SCHEMA_ID : undefined));
    ***REMOVED***
***REMOVED***
exports.Ajv2020 = Ajv2020;
module.exports = exports = Ajv2020;
module.exports.Ajv2020 = Ajv2020;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.default = Ajv2020;
var validate_1 = require("./compile/validate");
Object.defineProperty(exports, "KeywordCxt", { enumerable: true, get: function () { return validate_1.KeywordCxt; ***REMOVED*** ***REMOVED***);
var codegen_1 = require("./compile/codegen");
Object.defineProperty(exports, "_", { enumerable: true, get: function () { return codegen_1._; ***REMOVED*** ***REMOVED***);
Object.defineProperty(exports, "str", { enumerable: true, get: function () { return codegen_1.str; ***REMOVED*** ***REMOVED***);
Object.defineProperty(exports, "stringify", { enumerable: true, get: function () { return codegen_1.stringify; ***REMOVED*** ***REMOVED***);
Object.defineProperty(exports, "nil", { enumerable: true, get: function () { return codegen_1.nil; ***REMOVED*** ***REMOVED***);
Object.defineProperty(exports, "Name", { enumerable: true, get: function () { return codegen_1.Name; ***REMOVED*** ***REMOVED***);
Object.defineProperty(exports, "CodeGen", { enumerable: true, get: function () { return codegen_1.CodeGen; ***REMOVED*** ***REMOVED***);
var validation_error_1 = require("./runtime/validation_error");
Object.defineProperty(exports, "ValidationError", { enumerable: true, get: function () { return validation_error_1.default; ***REMOVED*** ***REMOVED***);
var ref_error_1 = require("./compile/ref_error");
Object.defineProperty(exports, "MissingRefError", { enumerable: true, get: function () { return ref_error_1.default; ***REMOVED*** ***REMOVED***);
