// enquirer/lib/ansi.js
'use strict';

const isTerm = process.env.TERM_PROGRAM === 'Apple_Terminal';
const stripAnsi = require('strip-ansi');
const utils = require('./utils');
const ansi = module.exports = exports;
const ESC = '\u001b[';
const BEL = '\u0007';
let hidden = false;

const code = ansi.code = {
  bell: BEL,
  beep: BEL,
  beginning: `${ESC***REMOVED***G`,
  down: `${ESC***REMOVED***J`,
  esc: ESC,
  getPosition: `${ESC***REMOVED***6n`,
  hide: `${ESC***REMOVED***?25l`,
  line: `${ESC***REMOVED***2K`,
  lineEnd: `${ESC***REMOVED***K`,
  lineStart: `${ESC***REMOVED***1K`,
  restorePosition: ESC + (isTerm ? '8' : 'u'),
  savePosition: ESC + (isTerm ? '7' : 's'),
  screen: `${ESC***REMOVED***2J`,
  show: `${ESC***REMOVED***?25h`,
  up: `${ESC***REMOVED***1J`
***REMOVED***;

const cursor = ansi.cursor = {
  get hidden() {
    return hidden;
  ***REMOVED***,

  hide() {
    hidden = true;
    return code.hide;
  ***REMOVED***,
  show() {
    hidden = false;
    return code.show;
  ***REMOVED***,

  forward: (count = 1) => `${ESC***REMOVED***${count***REMOVED***C`,
  backward: (count = 1) => `${ESC***REMOVED***${count***REMOVED***D`,
  nextLine: (count = 1) => `${ESC***REMOVED***E`.repeat(count),
  prevLine: (count = 1) => `${ESC***REMOVED***F`.repeat(count),

  up: (count = 1) => count ? `${ESC***REMOVED***${count***REMOVED***A` : '',
  down: (count = 1) => count ? `${ESC***REMOVED***${count***REMOVED***B` : '',
  right: (count = 1) => count ? `${ESC***REMOVED***${count***REMOVED***C` : '',
  left: (count = 1) => count ? `${ESC***REMOVED***${count***REMOVED***D` : '',

  to(x, y) {
    return y ? `${ESC***REMOVED***${y + 1***REMOVED***;${x + 1***REMOVED***H` : `${ESC***REMOVED***${x + 1***REMOVED***G`;
  ***REMOVED***,

  move(x = 0, y = 0) {
    let res = '';
    res += (x < 0) ? cursor.left(-x) : (x > 0) ? cursor.right(x) : '';
    res += (y < 0) ? cursor.up(-y) : (y > 0) ? cursor.down(y) : '';
    return res;
  ***REMOVED***,
  strLen(str) {
    // to suport chinese
    var realLength = 0, len = str.length, charCode = -1;
    for (var i = 0; i < len; i++) {
      charCode = str.charCodeAt(i);
      if (charCode >= 0 && charCode <= 128) realLength += 1;
      else realLength += 2;
    ***REMOVED***
    return realLength;
  ***REMOVED***,
  restore(state = {***REMOVED***) {
    let { after, cursor, initial, input, prompt, size, value ***REMOVED*** = state;
    initial = utils.isPrimitive(initial) ? String(initial) : '';
    input = utils.isPrimitive(input) ? String(input) : '';
    value = utils.isPrimitive(value) ? String(value) : '';

    if (size) {
      let codes = ansi.cursor.up(size) + ansi.cursor.to(this.strLen(prompt));
      let diff = input.length - cursor;
      if (diff > 0) {
        codes += ansi.cursor.left(diff);
      ***REMOVED***
      return codes;
    ***REMOVED***

    if (value || after) {
      let pos = (!input && !!initial) ? - this.strLen(initial) : -this.strLen(input) + cursor;
      if (after) pos -= this.strLen(after);
      if (input === '' && initial && !prompt.includes(initial)) {
        pos += this.strLen(initial);
      ***REMOVED***
      return ansi.cursor.move(pos);
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

const erase = ansi.erase = {
  screen: code.screen,
  up: code.up,
  down: code.down,
  line: code.line,
  lineEnd: code.lineEnd,
  lineStart: code.lineStart,
  lines(n) {
    let str = '';
    for (let i = 0; i < n; i++) {
      str += ansi.erase.line + (i < n - 1 ? ansi.cursor.up(1) : '');
    ***REMOVED***
    if (n) str += ansi.code.beginning;
    return str;
  ***REMOVED***
***REMOVED***;

ansi.clear = (input = '', columns = process.stdout.columns) => {
  if (!columns) return erase.line + cursor.to(0);
  let width = str => [...stripAnsi(str)].length;
  let lines = input.split(/\r?\n/);
  let rows = 0;
  for (let line of lines) {
    rows += 1 + Math.floor(Math.max(width(line) - 1, 0) / columns);
  ***REMOVED***
  return (erase.line + cursor.prevLine()).repeat(rows - 1) + erase.line + cursor.to(0);
***REMOVED***;
