// enquirer/lib/interpolate.js
'use strict';

const stripAnsi = require('strip-ansi');
const clean = (str = '') => {
  return typeof str === 'string' ? str.replace(/^['"]|['"]$/g, '') : '';
***REMOVED***;

/**
 * This file contains the interpolation and rendering logic for
 * the Snippet prompt.
 */

class Item {
  constructor(token) {
    this.name = token.key;
    this.field = token.field || {***REMOVED***;
    this.value = clean(token.initial || this.field.initial || '');
    this.message = token.message || this.name;
    this.cursor = 0;
    this.input = '';
    this.lines = [];
  ***REMOVED***
***REMOVED***

const tokenize = async(options = {***REMOVED***, defaults = {***REMOVED***, fn = token => token) => {
  let unique = new Set();
  let fields = options.fields || [];
  let input = options.template;
  let tabstops = [];
  let items = [];
  let keys = [];
  let line = 1;

  if (typeof input === 'function') {
    input = await input();
  ***REMOVED***

  let i = -1;
  let next = () => input[++i];
  let peek = () => input[i + 1];
  let push = token => {
    token.line = line;
    tabstops.push(token);
  ***REMOVED***;

  push({ type: 'bos', value: '' ***REMOVED***);

  while (i < input.length - 1) {
    let value = next();

    if (/^[^\S\n ]$/.test(value)) {
      push({ type: 'text', value ***REMOVED***);
      continue;
    ***REMOVED***

    if (value === '\n') {
      push({ type: 'newline', value ***REMOVED***);
      line++;
      continue;
    ***REMOVED***

    if (value === '\\') {
      value += next();
      push({ type: 'text', value ***REMOVED***);
      continue;
    ***REMOVED***

    if ((value === '$' || value === '#' || value === '{') && peek() === '{') {
      let n = next();
      value += n;

      let token = { type: 'template', open: value, inner: '', close: '', value ***REMOVED***;
      let ch;

      while ((ch = next())) {
        if (ch === '***REMOVED***') {
          if (peek() === '***REMOVED***') ch += next();
          token.value += ch;
          token.close = ch;
          break;
        ***REMOVED***

        if (ch === ':') {
          token.initial = '';
          token.key = token.inner;
        ***REMOVED*** else if (token.initial !== void 0) {
          token.initial += ch;
        ***REMOVED***

        token.value += ch;
        token.inner += ch;
      ***REMOVED***

      token.template = token.open + (token.initial || token.inner) + token.close;
      token.key = token.key || token.inner;

      if (hasOwnProperty.call(defaults, token.key)) {
        token.initial = defaults[token.key];
      ***REMOVED***

      token = fn(token);
      push(token);

      keys.push(token.key);
      unique.add(token.key);

      let item = items.find(item => item.name === token.key);
      token.field = fields.find(ch => ch.name === token.key);

      if (!item) {
        item = new Item(token);
        items.push(item);
      ***REMOVED***

      item.lines.push(token.line - 1);
      continue;
    ***REMOVED***

    let last = tabstops[tabstops.length - 1];
    if (last.type === 'text' && last.line === line) {
      last.value += value;
    ***REMOVED*** else {
      push({ type: 'text', value ***REMOVED***);
    ***REMOVED***
  ***REMOVED***

  push({ type: 'eos', value: '' ***REMOVED***);
  return { input, tabstops, unique, keys, items ***REMOVED***;
***REMOVED***;

module.exports = async prompt => {
  let options = prompt.options;
  let required = new Set(options.required === true ? [] : (options.required || []));
  let defaults = { ...options.values, ...options.initial ***REMOVED***;
  let { tabstops, items, keys ***REMOVED*** = await tokenize(options, defaults);

  let result = createFn('result', prompt, options);
  let format = createFn('format', prompt, options);
  let isValid = createFn('validate', prompt, options, true);
  let isVal = prompt.isValue.bind(prompt);

  return async(state = {***REMOVED***, submitted = false) => {
    let index = 0;

    state.required = required;
    state.items = items;
    state.keys = keys;
    state.output = '';

    let validate = async(value, state, item, index) => {
      let error = await isValid(value, state, item, index);
      if (error === false) {
        return 'Invalid field ' + item.name;
      ***REMOVED***
      return error;
    ***REMOVED***;

    for (let token of tabstops) {
      let value = token.value;
      let key = token.key;

      if (token.type !== 'template') {
        if (value) state.output += value;
        continue;
      ***REMOVED***

      if (token.type === 'template') {
        let item = items.find(ch => ch.name === key);

        if (options.required === true) {
          state.required.add(item.name);
        ***REMOVED***

        let val = [item.input, state.values[item.value], item.value, value].find(isVal);
        let field = item.field || {***REMOVED***;
        let message = field.message || token.inner;

        if (submitted) {
          let error = await validate(state.values[key], state, item, index);
          if ((error && typeof error === 'string') || error === false) {
            state.invalid.set(key, error);
            continue;
          ***REMOVED***

          state.invalid.delete(key);
          let res = await result(state.values[key], state, item, index);
          state.output += stripAnsi(res);
          continue;
        ***REMOVED***

        item.placeholder = false;

        let before = value;
        value = await format(value, state, item, index);

        if (val !== value) {
          state.values[key] = val;
          value = prompt.styles.typing(val);
          state.missing.delete(message);

        ***REMOVED*** else {
          state.values[key] = void 0;
          val = `<${message***REMOVED***>`;
          value = prompt.styles.primary(val);
          item.placeholder = true;

          if (state.required.has(key)) {
            state.missing.add(message);
          ***REMOVED***
        ***REMOVED***

        if (state.missing.has(message) && state.validating) {
          value = prompt.styles.warning(val);
        ***REMOVED***

        if (state.invalid.has(key) && state.validating) {
          value = prompt.styles.danger(val);
        ***REMOVED***

        if (index === state.index) {
          if (before !== value) {
            value = prompt.styles.underline(value);
          ***REMOVED*** else {
            value = prompt.styles.heading(stripAnsi(value));
          ***REMOVED***
        ***REMOVED***

        index++;
      ***REMOVED***

      if (value) {
        state.output += value;
      ***REMOVED***
    ***REMOVED***

    let lines = state.output.split('\n').map(l => ' ' + l);
    let len = items.length;
    let done = 0;

    for (let item of items) {
      if (state.invalid.has(item.name)) {
        item.lines.forEach(i => {
          if (lines[i][0] !== ' ') return;
          lines[i] = state.styles.danger(state.symbols.bullet) + lines[i].slice(1);
        ***REMOVED***);
      ***REMOVED***

      if (prompt.isValue(state.values[item.name])) {
        done++;
      ***REMOVED***
    ***REMOVED***

    state.completed = ((done / len) * 100).toFixed(0);
    state.output = lines.join('\n');
    return state.output;
  ***REMOVED***;
***REMOVED***;

function createFn(prop, prompt, options, fallback) {
  return (value, state, item, index) => {
    if (typeof item.field[prop] === 'function') {
      return item.field[prop].call(prompt, value, state, item, index);
    ***REMOVED***
    return [fallback, value].find(v => prompt.isValue(v));
  ***REMOVED***;
***REMOVED***
