// enquirer/lib/roles.js
'use strict';

const utils = require('./utils');
const roles = {
  default(prompt, choice) {
    return choice;
  ***REMOVED***,
  checkbox(prompt, choice) {
    throw new Error('checkbox role is not implemented yet');
  ***REMOVED***,
  editable(prompt, choice) {
    throw new Error('editable role is not implemented yet');
  ***REMOVED***,
  expandable(prompt, choice) {
    throw new Error('expandable role is not implemented yet');
  ***REMOVED***,
  heading(prompt, choice) {
    choice.disabled = '';
    choice.indicator = [choice.indicator, ' '].find(v => v != null);
    choice.message = choice.message || '';
    return choice;
  ***REMOVED***,
  input(prompt, choice) {
    throw new Error('input role is not implemented yet');
  ***REMOVED***,
  option(prompt, choice) {
    return roles.default(prompt, choice);
  ***REMOVED***,
  radio(prompt, choice) {
    throw new Error('radio role is not implemented yet');
  ***REMOVED***,
  separator(prompt, choice) {
    choice.disabled = '';
    choice.indicator = [choice.indicator, ' '].find(v => v != null);
    choice.message = choice.message || prompt.symbols.line.repeat(5);
    return choice;
  ***REMOVED***,
  spacer(prompt, choice) {
    return choice;
  ***REMOVED***
***REMOVED***;

module.exports = (name, options = {***REMOVED***) => {
  let role = utils.merge({***REMOVED***, roles, options.roles);
  return role[name] || role.default;
***REMOVED***;
