// enquirer/lib/styles.js
'use strict';

const utils = require('./utils');
const colors = require('ansi-colors');

const styles = {
  default: colors.noop,
  noop: colors.noop,

  /**
   * Modifiers
   */

  set inverse(custom) {
    this._inverse = custom;
  ***REMOVED***,
  get inverse() {
    return this._inverse || utils.inverse(this.primary);
  ***REMOVED***,

  set complement(custom) {
    this._complement = custom;
  ***REMOVED***,
  get complement() {
    return this._complement || utils.complement(this.primary);
  ***REMOVED***,

  /**
   * Main color
   */

  primary: colors.cyan,

  /**
   * Main palette
   */

  success: colors.green,
  danger: colors.magenta,
  strong: colors.bold,
  warning: colors.yellow,
  muted: colors.dim,
  disabled: colors.gray,
  dark: colors.dim.gray,
  underline: colors.underline,

  set info(custom) {
    this._info = custom;
  ***REMOVED***,
  get info() {
    return this._info || this.primary;
  ***REMOVED***,

  set em(custom) {
    this._em = custom;
  ***REMOVED***,
  get em() {
    return this._em || this.primary.underline;
  ***REMOVED***,

  set heading(custom) {
    this._heading = custom;
  ***REMOVED***,
  get heading() {
    return this._heading || this.muted.underline;
  ***REMOVED***,

  /**
   * Statuses
   */

  set pending(custom) {
    this._pending = custom;
  ***REMOVED***,
  get pending() {
    return this._pending || this.primary;
  ***REMOVED***,

  set submitted(custom) {
    this._submitted = custom;
  ***REMOVED***,
  get submitted() {
    return this._submitted || this.success;
  ***REMOVED***,

  set cancelled(custom) {
    this._cancelled = custom;
  ***REMOVED***,
  get cancelled() {
    return this._cancelled || this.danger;
  ***REMOVED***,

  /**
   * Special styling
   */

  set typing(custom) {
    this._typing = custom;
  ***REMOVED***,
  get typing() {
    return this._typing || this.dim;
  ***REMOVED***,

  set placeholder(custom) {
    this._placeholder = custom;
  ***REMOVED***,
  get placeholder() {
    return this._placeholder || this.primary.dim;
  ***REMOVED***,

  set highlight(custom) {
    this._highlight = custom;
  ***REMOVED***,
  get highlight() {
    return this._highlight || this.inverse;
  ***REMOVED***
***REMOVED***;

styles.merge = (options = {***REMOVED***) => {
  if (options.styles && typeof options.styles.enabled === 'boolean') {
    colors.enabled = options.styles.enabled;
  ***REMOVED***
  if (options.styles && typeof options.styles.visible === 'boolean') {
    colors.visible = options.styles.visible;
  ***REMOVED***

  let result = utils.merge({***REMOVED***, styles, options.styles);
  delete result.merge;

  for (let key of Object.keys(colors)) {
    if (!hasOwnProperty.call(result, key)) {
      Reflect.defineProperty(result, key, { get: () => colors[key] ***REMOVED***);
    ***REMOVED***
  ***REMOVED***

  for (let key of Object.keys(colors.styles)) {
    if (!hasOwnProperty.call(result, key)) {
      Reflect.defineProperty(result, key, { get: () => colors[key] ***REMOVED***);
    ***REMOVED***
  ***REMOVED***
  return result;
***REMOVED***;

module.exports = styles;
