// lilconfig/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.lilconfigSync = exports.lilconfig = exports.defaultLoaders = void 0;
const path = require("path");
const fs = require("fs");
const os = require("os");
const fsReadFileAsync = fs.promises.readFile;
function getDefaultSearchPlaces(name) {
    return [
        'package.json',
        `.${name***REMOVED***rc.json`,
        `.${name***REMOVED***rc.js`,
        `.${name***REMOVED***rc.cjs`,
        `.config/${name***REMOVED***rc`,
        `.config/${name***REMOVED***rc.json`,
        `.config/${name***REMOVED***rc.js`,
        `.config/${name***REMOVED***rc.cjs`,
        `${name***REMOVED***.config.js`,
        `${name***REMOVED***.config.cjs`,
    ];
***REMOVED***
function getSearchPaths(startDir, stopDir) {
    return startDir
        .split(path.sep)
        .reduceRight((acc, _, ind, arr) => {
        const currentPath = arr.slice(0, ind + 1).join(path.sep);
        if (!acc.passedStopDir)
            acc.searchPlaces.push(currentPath || path.sep);
        if (currentPath === stopDir)
            acc.passedStopDir = true;
        return acc;
    ***REMOVED***, { searchPlaces: [], passedStopDir: false ***REMOVED***).searchPlaces;
***REMOVED***
exports.defaultLoaders = Object.freeze({
    '.js': require,
    '.json': require,
    '.cjs': require,
    noExt(_, content) {
        return JSON.parse(content);
    ***REMOVED***,
***REMOVED***);
function getExtDesc(ext) {
    return ext === 'noExt' ? 'files without extensions' : `extension "${ext***REMOVED***"`;
***REMOVED***
function getOptions(name, options = {***REMOVED***) {
    const conf = {
        stopDir: os.homedir(),
        searchPlaces: getDefaultSearchPlaces(name),
        ignoreEmptySearchPlaces: true,
        transform: (x) => x,
        packageProp: [name],
        ...options,
        loaders: { ...exports.defaultLoaders, ...options.loaders ***REMOVED***,
    ***REMOVED***;
    conf.searchPlaces.forEach(place => {
        const key = path.extname(place) || 'noExt';
        const loader = conf.loaders[key];
        if (!loader) {
            throw new Error(`No loader specified for ${getExtDesc(key)***REMOVED***, so searchPlaces item "${place***REMOVED***" is invalid`);
        ***REMOVED***
        if (typeof loader !== 'function') {
            throw new Error(`loader for ${getExtDesc(key)***REMOVED*** is not a function (type provided: "${typeof loader***REMOVED***"), so searchPlaces item "${place***REMOVED***" is invalid`);
        ***REMOVED***
    ***REMOVED***);
    return conf;
***REMOVED***
function getPackageProp(props, obj) {
    if (typeof props === 'string' && props in obj)
        return obj[props];
    return ((Array.isArray(props) ? props : props.split('.')).reduce((acc, prop) => (acc === undefined ? acc : acc[prop]), obj) || null);
***REMOVED***
function getSearchItems(searchPlaces, searchPaths) {
    return searchPaths.reduce((acc, searchPath) => {
        searchPlaces.forEach(sp => acc.push({
            searchPlace: sp,
            filepath: path.join(searchPath, sp),
            loaderKey: path.extname(sp) || 'noExt',
        ***REMOVED***));
        return acc;
    ***REMOVED***, []);
***REMOVED***
function validateFilePath(filepath) {
    if (!filepath)
        throw new Error('load must pass a non-empty string');
***REMOVED***
function validateLoader(loader, ext) {
    if (!loader)
        throw new Error(`No loader specified for extension "${ext***REMOVED***"`);
    if (typeof loader !== 'function')
        throw new Error('loader is not a function');
***REMOVED***
function lilconfig(name, options) {
    const { ignoreEmptySearchPlaces, loaders, packageProp, searchPlaces, stopDir, transform, ***REMOVED*** = getOptions(name, options);
    return {
        async search(searchFrom = process.cwd()) {
            const searchPaths = getSearchPaths(searchFrom, stopDir);
            const result = {
                config: null,
                filepath: '',
            ***REMOVED***;
            const searchItems = getSearchItems(searchPlaces, searchPaths);
            for (const { searchPlace, filepath, loaderKey ***REMOVED*** of searchItems) {
                try {
                    await fs.promises.access(filepath);
                ***REMOVED***
                catch (_a) {
                    continue;
                ***REMOVED***
                const content = String(await fsReadFileAsync(filepath));
                const loader = loaders[loaderKey];
                if (searchPlace === 'package.json') {
                    const pkg = await loader(filepath, content);
                    const maybeConfig = getPackageProp(packageProp, pkg);
                    if (maybeConfig != null) {
                        result.config = maybeConfig;
                        result.filepath = filepath;
                        break;
                    ***REMOVED***
                    continue;
                ***REMOVED***
                const isEmpty = content.trim() === '';
                if (isEmpty && ignoreEmptySearchPlaces)
                    continue;
                if (isEmpty) {
                    result.isEmpty = true;
                    result.config = undefined;
                ***REMOVED***
                else {
                    validateLoader(loader, loaderKey);
                    result.config = await loader(filepath, content);
                ***REMOVED***
                result.filepath = filepath;
                break;
            ***REMOVED***
            if (result.filepath === '' && result.config === null)
                return transform(null);
            return transform(result);
        ***REMOVED***,
        async load(filepath) {
            validateFilePath(filepath);
            const absPath = path.resolve(process.cwd(), filepath);
            const { base, ext ***REMOVED*** = path.parse(absPath);
            const loaderKey = ext || 'noExt';
            const loader = loaders[loaderKey];
            validateLoader(loader, loaderKey);
            const content = String(await fsReadFileAsync(absPath));
            if (base === 'package.json') {
                const pkg = await loader(absPath, content);
                return transform({
                    config: getPackageProp(packageProp, pkg),
                    filepath: absPath,
                ***REMOVED***);
            ***REMOVED***
            const result = {
                config: null,
                filepath: absPath,
            ***REMOVED***;
            const isEmpty = content.trim() === '';
            if (isEmpty && ignoreEmptySearchPlaces)
                return transform({
                    config: undefined,
                    filepath: absPath,
                    isEmpty: true,
                ***REMOVED***);
            result.config = isEmpty
                ? undefined
                : await loader(absPath, content);
            return transform(isEmpty ? { ...result, isEmpty, config: undefined ***REMOVED*** : result);
        ***REMOVED***,
    ***REMOVED***;
***REMOVED***
exports.lilconfig = lilconfig;
function lilconfigSync(name, options) {
    const { ignoreEmptySearchPlaces, loaders, packageProp, searchPlaces, stopDir, transform, ***REMOVED*** = getOptions(name, options);
    return {
        search(searchFrom = process.cwd()) {
            const searchPaths = getSearchPaths(searchFrom, stopDir);
            const result = {
                config: null,
                filepath: '',
            ***REMOVED***;
            const searchItems = getSearchItems(searchPlaces, searchPaths);
            for (const { searchPlace, filepath, loaderKey ***REMOVED*** of searchItems) {
                try {
                    fs.accessSync(filepath);
                ***REMOVED***
                catch (_a) {
                    continue;
                ***REMOVED***
                const loader = loaders[loaderKey];
                const content = String(fs.readFileSync(filepath));
                if (searchPlace === 'package.json') {
                    const pkg = loader(filepath, content);
                    const maybeConfig = getPackageProp(packageProp, pkg);
                    if (maybeConfig != null) {
                        result.config = maybeConfig;
                        result.filepath = filepath;
                        break;
                    ***REMOVED***
                    continue;
                ***REMOVED***
                const isEmpty = content.trim() === '';
                if (isEmpty && ignoreEmptySearchPlaces)
                    continue;
                if (isEmpty) {
                    result.isEmpty = true;
                    result.config = undefined;
                ***REMOVED***
                else {
                    validateLoader(loader, loaderKey);
                    result.config = loader(filepath, content);
                ***REMOVED***
                result.filepath = filepath;
                break;
            ***REMOVED***
            if (result.filepath === '' && result.config === null)
                return transform(null);
            return transform(result);
        ***REMOVED***,
        load(filepath) {
            validateFilePath(filepath);
            const absPath = path.resolve(process.cwd(), filepath);
            const { base, ext ***REMOVED*** = path.parse(absPath);
            const loaderKey = ext || 'noExt';
            const loader = loaders[loaderKey];
            validateLoader(loader, loaderKey);
            const content = String(fs.readFileSync(absPath));
            if (base === 'package.json') {
                const pkg = loader(absPath, content);
                return transform({
                    config: getPackageProp(packageProp, pkg),
                    filepath: absPath,
                ***REMOVED***);
            ***REMOVED***
            const result = {
                config: null,
                filepath: absPath,
            ***REMOVED***;
            const isEmpty = content.trim() === '';
            if (isEmpty && ignoreEmptySearchPlaces)
                return transform({
                    filepath: absPath,
                    config: undefined,
                    isEmpty: true,
                ***REMOVED***);
            result.config = isEmpty ? undefined : loader(absPath, content);
            return transform(isEmpty ? { ...result, isEmpty, config: undefined ***REMOVED*** : result);
        ***REMOVED***,
    ***REMOVED***;
***REMOVED***
exports.lilconfigSync = lilconfigSync;
