// jsdom/lib/jsdom/utils.js
"use strict";
const path = require("path");
const whatwgURL = require("whatwg-url");
const { domSymbolTree ***REMOVED*** = require("./living/helpers/internal-constants");
const SYMBOL_TREE_POSITION = require("symbol-tree").TreePosition;

exports.hasWeakRefs = typeof WeakRef === "function";

exports.toFileUrl = function (fileName) {
  // Beyond just the `path.resolve`, this is mostly for the benefit of Windows,
  // where we need to convert "\" to "/" and add an extra "/" prefix before the
  // drive letter.
  let pathname = path.resolve(process.cwd(), fileName).replace(/\\/g, "/");
  if (pathname[0] !== "/") {
    pathname = "/" + pathname;
  ***REMOVED***

  // path might contain spaces, so convert those to %20
  return "file://" + encodeURI(pathname);
***REMOVED***;

/**
 * Define a set of properties on an object, by copying the property descriptors
 * from the original object.
 *
 * - `object` {Object***REMOVED*** the target object
 * - `properties` {Object***REMOVED*** the source from which to copy property descriptors
 */
exports.define = function define(object, properties) {
  for (const name of Object.getOwnPropertyNames(properties)) {
    const propDesc = Object.getOwnPropertyDescriptor(properties, name);
    Object.defineProperty(object, name, propDesc);
  ***REMOVED***
***REMOVED***;

/**
 * Define a list of constants on a constructor and its .prototype
 *
 * - `Constructor` {Function***REMOVED*** the constructor to define the constants on
 * - `propertyMap` {Object***REMOVED***  key/value map of properties to define
 */
exports.addConstants = function addConstants(Constructor, propertyMap) {
  for (const property in propertyMap) {
    const value = propertyMap[property];
    addConstant(Constructor, property, value);
    addConstant(Constructor.prototype, property, value);
  ***REMOVED***
***REMOVED***;

function addConstant(object, property, value) {
  Object.defineProperty(object, property, {
    configurable: false,
    enumerable: true,
    writable: false,
    value
  ***REMOVED***);
***REMOVED***

exports.mixin = (target, source) => {
  const keys = Reflect.ownKeys(source);
  for (let i = 0; i < keys.length; ++i) {
    if (keys[i] in target) {
      continue;
    ***REMOVED***

    Object.defineProperty(target, keys[i], Object.getOwnPropertyDescriptor(source, keys[i]));
  ***REMOVED***
***REMOVED***;

let memoizeQueryTypeCounter = 0;

/**
 * Returns a version of a method that memoizes specific types of calls on the object
 *
 * - `fn` {Function***REMOVED*** the method to be memozied
 */
exports.memoizeQuery = function memoizeQuery(fn) {
  // Only memoize query functions with arity <= 2
  if (fn.length > 2) {
    return fn;
  ***REMOVED***

  const type = memoizeQueryTypeCounter++;

  return function (...args) {
    if (!this._memoizedQueries) {
      return fn.apply(this, args);
    ***REMOVED***

    if (!this._memoizedQueries[type]) {
      this._memoizedQueries[type] = Object.create(null);
    ***REMOVED***

    let key;
    if (args.length === 1 && typeof args[0] === "string") {
      key = args[0];
    ***REMOVED*** else if (args.length === 2 && typeof args[0] === "string" && typeof args[1] === "string") {
      key = args[0] + "::" + args[1];
    ***REMOVED*** else {
      return fn.apply(this, args);
    ***REMOVED***

    if (!(key in this._memoizedQueries[type])) {
      this._memoizedQueries[type][key] = fn.apply(this, args);
    ***REMOVED***
    return this._memoizedQueries[type][key];
  ***REMOVED***;
***REMOVED***;

function isValidAbsoluteURL(str) {
  return whatwgURL.parseURL(str) !== null;
***REMOVED***

exports.isValidTargetOrigin = function (str) {
  return str === "*" || str === "/" || isValidAbsoluteURL(str);
***REMOVED***;

exports.simultaneousIterators = function* (first, second) {
  for (;;) {
    const firstResult = first.next();
    const secondResult = second.next();

    if (firstResult.done && secondResult.done) {
      return;
    ***REMOVED***

    yield [
      firstResult.done ? null : firstResult.value,
      secondResult.done ? null : secondResult.value
    ];
  ***REMOVED***
***REMOVED***;

exports.treeOrderSorter = function (a, b) {
  const compare = domSymbolTree.compareTreePosition(a, b);

  if (compare & SYMBOL_TREE_POSITION.PRECEDING) { // b is preceding a
    return 1;
  ***REMOVED***

  if (compare & SYMBOL_TREE_POSITION.FOLLOWING) {
    return -1;
  ***REMOVED***

  // disconnected or equal:
  return 0;
***REMOVED***;

/* eslint-disable global-require */

exports.Canvas = null;
let canvasInstalled = false;
try {
  require.resolve("canvas");
  canvasInstalled = true;
***REMOVED*** catch (e) {
  // canvas is not installed
***REMOVED***
if (canvasInstalled) {
  const Canvas = require("canvas");
  if (typeof Canvas.createCanvas === "function") {
    // In browserify, the require will succeed but return an empty object
    exports.Canvas = Canvas;
  ***REMOVED***
***REMOVED***
