// http-proxy-middleware/dist/config-factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.createConfig = void 0;
const isPlainObj = require("is-plain-obj");
const url = require("url");
const errors_1 = require("./errors");
const logger_1 = require("./logger");
const logger = (0, logger_1.getInstance)();
function createConfig(context, opts) {
    // structure of config object to be returned
    const config = {
        context: undefined,
        options: {***REMOVED***,
    ***REMOVED***;
    // app.use('/api', proxy({target:'http://localhost:9000'***REMOVED***));
    if (isContextless(context, opts)) {
        config.context = '/';
        config.options = Object.assign(config.options, context);
        // app.use('/api', proxy('http://localhost:9000'));
        // app.use(proxy('http://localhost:9000/api'));
    ***REMOVED***
    else if (isStringShortHand(context)) {
        const oUrl = url.parse(context);
        const target = [oUrl.protocol, '//', oUrl.host].join('');
        config.context = oUrl.pathname || '/';
        config.options = Object.assign(config.options, { target ***REMOVED***, opts);
        if (oUrl.protocol === 'ws:' || oUrl.protocol === 'wss:') {
            config.options.ws = true;
        ***REMOVED***
        // app.use('/api', proxy({target:'http://localhost:9000'***REMOVED***));
    ***REMOVED***
    else {
        config.context = context;
        config.options = Object.assign(config.options, opts);
    ***REMOVED***
    configureLogger(config.options);
    if (!config.options.target && !config.options.router) {
        throw new Error(errors_1.ERRORS.ERR_CONFIG_FACTORY_TARGET_MISSING);
    ***REMOVED***
    return config;
***REMOVED***
exports.createConfig = createConfig;
/**
 * Checks if a String only target/config is provided.
 * This can be just the host or with the optional path.
 *
 * @example
 *      app.use('/api', proxy('http://localhost:9000'));
 *      app.use(proxy('http://localhost:9000/api'));
 *
 * @param  {String***REMOVED***  context [description]
 * @return {Boolean***REMOVED***         [description]
 */
function isStringShortHand(context) {
    if (typeof context === 'string') {
        return !!url.parse(context).host;
    ***REMOVED***
***REMOVED***
/**
 * Checks if a Object only config is provided, without a context.
 * In this case the all paths will be proxied.
 *
 * @example
 *     app.use('/api', proxy({target:'http://localhost:9000'***REMOVED***));
 *
 * @param  {Object***REMOVED***  context [description]
 * @param  {****REMOVED***       opts    [description]
 * @return {Boolean***REMOVED***         [description]
 */
function isContextless(context, opts) {
    return isPlainObj(context) && (opts == null || Object.keys(opts).length === 0);
***REMOVED***
function configureLogger(options) {
    if (options.logLevel) {
        logger.setLevel(options.logLevel);
    ***REMOVED***
    if (options.logProvider) {
        logger.setProvider(options.logProvider);
    ***REMOVED***
***REMOVED***
