// http-proxy-middleware/dist/context-matcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.match = void 0;
const isGlob = require("is-glob");
const micromatch = require("micromatch");
const url = require("url");
const errors_1 = require("./errors");
function match(context, uri, req) {
    // single path
    if (isStringPath(context)) {
        return matchSingleStringPath(context, uri);
    ***REMOVED***
    // single glob path
    if (isGlobPath(context)) {
        return matchSingleGlobPath(context, uri);
    ***REMOVED***
    // multi path
    if (Array.isArray(context)) {
        if (context.every(isStringPath)) {
            return matchMultiPath(context, uri);
        ***REMOVED***
        if (context.every(isGlobPath)) {
            return matchMultiGlobPath(context, uri);
        ***REMOVED***
        throw new Error(errors_1.ERRORS.ERR_CONTEXT_MATCHER_INVALID_ARRAY);
    ***REMOVED***
    // custom matching
    if (typeof context === 'function') {
        const pathname = getUrlPathName(uri);
        return context(pathname, req);
    ***REMOVED***
    throw new Error(errors_1.ERRORS.ERR_CONTEXT_MATCHER_GENERIC);
***REMOVED***
exports.match = match;
/**
 * @param  {String***REMOVED*** context '/api'
 * @param  {String***REMOVED*** uri     'http://example.org/api/b/c/d.html'
 * @return {Boolean***REMOVED***
 */
function matchSingleStringPath(context, uri) {
    const pathname = getUrlPathName(uri);
    return pathname.indexOf(context) === 0;
***REMOVED***
function matchSingleGlobPath(pattern, uri) {
    const pathname = getUrlPathName(uri);
    const matches = micromatch([pathname], pattern);
    return matches && matches.length > 0;
***REMOVED***
function matchMultiGlobPath(patternList, uri) {
    return matchSingleGlobPath(patternList, uri);
***REMOVED***
/**
 * @param  {String***REMOVED*** contextList ['/api', '/ajax']
 * @param  {String***REMOVED*** uri     'http://example.org/api/b/c/d.html'
 * @return {Boolean***REMOVED***
 */
function matchMultiPath(contextList, uri) {
    let isMultiPath = false;
    for (const context of contextList) {
        if (matchSingleStringPath(context, uri)) {
            isMultiPath = true;
            break;
        ***REMOVED***
    ***REMOVED***
    return isMultiPath;
***REMOVED***
/**
 * Parses URI and returns RFC 3986 path
 *
 * @param  {String***REMOVED*** uri from req.url
 * @return {String***REMOVED***     RFC 3986 path
 */
function getUrlPathName(uri) {
    return uri && url.parse(uri).pathname;
***REMOVED***
function isStringPath(context) {
    return typeof context === 'string' && !isGlob(context);
***REMOVED***
function isGlobPath(context) {
    return isGlob(context);
***REMOVED***
