// http-proxy-middleware/dist/logger.js
"use strict";
/* eslint-disable prefer-rest-params */
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.getArrow = exports.getInstance = void 0;
const util = require("util");
let loggerInstance;
const defaultProvider = {
    // tslint:disable: no-console
    log: console.log,
    debug: console.log,
    info: console.info,
    warn: console.warn,
    error: console.error,
***REMOVED***;
// log level 'weight'
var LEVELS;
(function (LEVELS) {
    LEVELS[LEVELS["debug"] = 10] = "debug";
    LEVELS[LEVELS["info"] = 20] = "info";
    LEVELS[LEVELS["warn"] = 30] = "warn";
    LEVELS[LEVELS["error"] = 50] = "error";
    LEVELS[LEVELS["silent"] = 80] = "silent";
***REMOVED***)(LEVELS || (LEVELS = {***REMOVED***));
function getInstance() {
    if (!loggerInstance) {
        loggerInstance = new Logger();
    ***REMOVED***
    return loggerInstance;
***REMOVED***
exports.getInstance = getInstance;
class Logger {
    constructor() {
        this.setLevel('info');
        this.setProvider(() => defaultProvider);
    ***REMOVED***
    // log will log messages, regardless of logLevels
    log() {
        this.provider.log(this._interpolate.apply(null, arguments));
    ***REMOVED***
    debug() {
        if (this._showLevel('debug')) {
            this.provider.debug(this._interpolate.apply(null, arguments));
        ***REMOVED***
    ***REMOVED***
    info() {
        if (this._showLevel('info')) {
            this.provider.info(this._interpolate.apply(null, arguments));
        ***REMOVED***
    ***REMOVED***
    warn() {
        if (this._showLevel('warn')) {
            this.provider.warn(this._interpolate.apply(null, arguments));
        ***REMOVED***
    ***REMOVED***
    error() {
        if (this._showLevel('error')) {
            this.provider.error(this._interpolate.apply(null, arguments));
        ***REMOVED***
    ***REMOVED***
    setLevel(v) {
        if (this.isValidLevel(v)) {
            this.logLevel = v;
        ***REMOVED***
    ***REMOVED***
    setProvider(fn) {
        if (fn && this.isValidProvider(fn)) {
            this.provider = fn(defaultProvider);
        ***REMOVED***
    ***REMOVED***
    isValidProvider(fnProvider) {
        const result = true;
        if (fnProvider && typeof fnProvider !== 'function') {
            throw new Error('[HPM] Log provider config error. Expecting a function.');
        ***REMOVED***
        return result;
    ***REMOVED***
    isValidLevel(levelName) {
        const validLevels = Object.keys(LEVELS);
        const isValid = validLevels.includes(levelName);
        if (!isValid) {
            throw new Error('[HPM] Log level error. Invalid logLevel.');
        ***REMOVED***
        return isValid;
    ***REMOVED***
    /**
     * Decide to log or not to log, based on the log levels 'weight'
     * @param  {String***REMOVED***  showLevel [debug, info, warn, error, silent]
     * @return {Boolean***REMOVED***
     */
    _showLevel(showLevel) {
        let result = false;
        const currentLogLevel = LEVELS[this.logLevel];
        if (currentLogLevel && currentLogLevel <= LEVELS[showLevel]) {
            result = true;
        ***REMOVED***
        return result;
    ***REMOVED***
    // make sure logged messages and its data are return interpolated
    // make it possible for additional log data, such date/time or custom prefix.
    _interpolate(format, ...args) {
        const result = util.format(format, ...args);
        return result;
    ***REMOVED***
***REMOVED***
/**
 * -> normal proxy
 * => router
 * ~> pathRewrite
 * ≈> router + pathRewrite
 *
 * @param  {String***REMOVED*** originalPath
 * @param  {String***REMOVED*** newPath
 * @param  {String***REMOVED*** originalTarget
 * @param  {String***REMOVED*** newTarget
 * @return {String***REMOVED***
 */
function getArrow(originalPath, newPath, originalTarget, newTarget) {
    const arrow = ['>'];
    const isNewTarget = originalTarget !== newTarget; // router
    const isNewPath = originalPath !== newPath; // pathRewrite
    if (isNewPath && !isNewTarget) {
        arrow.unshift('~');
    ***REMOVED***
    else if (!isNewPath && isNewTarget) {
        arrow.unshift('=');
    ***REMOVED***
    else if (isNewPath && isNewTarget) {
        arrow.unshift('≈');
    ***REMOVED***
    else {
        arrow.unshift('-');
    ***REMOVED***
    return arrow.join('');
***REMOVED***
exports.getArrow = getArrow;
