// eslint-plugin-react/lib/rules/boolean-prop-naming.js
/**
 * @fileoverview Enforces consistent naming for boolean props
 * @author Ev Haus
 */

'use strict';

const flatMap = require('array.prototype.flatmap');
const values = require('object.values');

const Components = require('../util/Components');
const propsUtil = require('../util/props');
const astUtil = require('../util/ast');
const docsUrl = require('../util/docsUrl');
const propWrapperUtil = require('../util/propWrapper');
const report = require('../util/report');
const eslintUtil = require('../util/eslint');

const getSourceCode = eslintUtil.getSourceCode;
const getText = eslintUtil.getText;

/**
 * Checks if prop is nested
 * @param {Object***REMOVED*** prop Property object, single prop type declaration
 * @returns {boolean***REMOVED***
 */
function nestedPropTypes(prop) {
  return (
    prop.type === 'Property'
    && astUtil.isCallExpression(prop.value)
  );
***REMOVED***

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

const messages = {
  patternMismatch: 'Prop name `{{propName***REMOVED******REMOVED***` doesn’t match rule `{{pattern***REMOVED******REMOVED***`',
***REMOVED***;

/** @type {import('eslint').Rule.RuleModule***REMOVED*** */
module.exports = {
  meta: {
    docs: {
      category: 'Stylistic Issues',
      description: 'Enforces consistent naming for boolean props',
      recommended: false,
      url: docsUrl('boolean-prop-naming'),
    ***REMOVED***,

    messages,

    schema: [{
      additionalProperties: false,
      properties: {
        propTypeNames: {
          items: {
            type: 'string',
          ***REMOVED***,
          minItems: 1,
          type: 'array',
          uniqueItems: true,
        ***REMOVED***,
        rule: {
          default: '^(is|has)[A-Z]([A-Za-z0-9]?)+',
          minLength: 1,
          type: 'string',
        ***REMOVED***,
        message: {
          minLength: 1,
          type: 'string',
        ***REMOVED***,
        validateNested: {
          default: false,
          type: 'boolean',
        ***REMOVED***,
      ***REMOVED***,
      type: 'object',
    ***REMOVED***],
  ***REMOVED***,

  create: Components.detect((context, components, utils) => {
    const config = context.options[0] || {***REMOVED***;
    const rule = config.rule ? new RegExp(config.rule) : null;
    const propTypeNames = config.propTypeNames || ['bool'];

    // Remembers all Flowtype object definitions
    const objectTypeAnnotations = new Map();

    /**
     * Returns the prop key to ensure we handle the following cases:
     * propTypes: {
     *   full: React.PropTypes.bool,
     *   short: PropTypes.bool,
     *   direct: bool,
     *   required: PropTypes.bool.isRequired
     * ***REMOVED***
     * @param {Object***REMOVED*** node The node we're getting the name of
     * @returns {string | null***REMOVED***
     */
    function getPropKey(node) {
      // Check for `ExperimentalSpreadProperty` (eslint 3/4) and `SpreadElement` (eslint 5)
      // so we can skip validation of those fields.
      // Otherwise it will look for `node.value.property` which doesn't exist and breaks eslint.
      if (node.type === 'ExperimentalSpreadProperty' || node.type === 'SpreadElement') {
        return null;
      ***REMOVED***
      if (node.value && node.value.property) {
        const name = node.value.property.name;
        if (name === 'isRequired') {
          if (node.value.object && node.value.object.property) {
            return node.value.object.property.name;
          ***REMOVED***
          return null;
        ***REMOVED***
        return name;
      ***REMOVED***
      if (node.value && node.value.type === 'Identifier') {
        return node.value.name;
      ***REMOVED***
      return null;
    ***REMOVED***

    /**
     * Returns the name of the given node (prop)
     * @param {Object***REMOVED*** node The node we're getting the name of
     * @returns {string***REMOVED***
     */
    function getPropName(node) {
      // Due to this bug https://github.com/babel/babel-eslint/issues/307
      // we can't get the name of the Flow object key name. So we have
      // to hack around it for now.
      if (node.type === 'ObjectTypeProperty') {
        return getSourceCode(context).getFirstToken(node).value;
      ***REMOVED***

      return node.key.name;
    ***REMOVED***

    /**
     * Checks if prop is declared in flow way
     * @param {Object***REMOVED*** prop Property object, single prop type declaration
     * @returns {boolean***REMOVED***
     */
    function flowCheck(prop) {
      return (
        prop.type === 'ObjectTypeProperty'
        && prop.value.type === 'BooleanTypeAnnotation'
        && rule.test(getPropName(prop)) === false
      );
    ***REMOVED***

    /**
     * Checks if prop is declared in regular way
     * @param {Object***REMOVED*** prop Property object, single prop type declaration
     * @returns {boolean***REMOVED***
     */
    function regularCheck(prop) {
      const propKey = getPropKey(prop);
      return (
        propKey
        && propTypeNames.indexOf(propKey) >= 0
        && rule.test(getPropName(prop)) === false
      );
    ***REMOVED***

    function tsCheck(prop) {
      if (prop.type !== 'TSPropertySignature') return false;
      const typeAnnotation = (prop.typeAnnotation || {***REMOVED***).typeAnnotation;
      return (
        typeAnnotation
        && typeAnnotation.type === 'TSBooleanKeyword'
        && rule.test(getPropName(prop)) === false
      );
    ***REMOVED***

    /**
     * Runs recursive check on all proptypes
     * @param {Array***REMOVED*** proptypes A list of Property object (for each proptype defined)
     * @param {Function***REMOVED*** addInvalidProp callback to run for each error
     */
    function runCheck(proptypes, addInvalidProp) {
      if (proptypes) {
        proptypes.forEach((prop) => {
          if (config.validateNested && nestedPropTypes(prop)) {
            runCheck(prop.value.arguments[0].properties, addInvalidProp);
            return;
          ***REMOVED***
          if (flowCheck(prop) || regularCheck(prop) || tsCheck(prop)) {
            addInvalidProp(prop);
          ***REMOVED***
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***

    /**
     * Checks and mark props with invalid naming
     * @param {Object***REMOVED*** node The component node we're testing
     * @param {Array***REMOVED*** proptypes A list of Property object (for each proptype defined)
     */
    function validatePropNaming(node, proptypes) {
      const component = components.get(node) || node;
      const invalidProps = component.invalidProps || [];

      runCheck(proptypes, (prop) => {
        invalidProps.push(prop);
      ***REMOVED***);

      components.set(node, {
        invalidProps,
      ***REMOVED***);
    ***REMOVED***

    /**
     * Reports invalid prop naming
     * @param {Object***REMOVED*** component The component to process
     */
    function reportInvalidNaming(component) {
      component.invalidProps.forEach((propNode) => {
        const propName = getPropName(propNode);
        report(context, config.message || messages.patternMismatch, !config.message && 'patternMismatch', {
          node: propNode,
          data: {
            component: propName,
            propName,
            pattern: config.rule,
          ***REMOVED***,
        ***REMOVED***);
      ***REMOVED***);
    ***REMOVED***

    function checkPropWrapperArguments(node, args) {
      if (!node || !Array.isArray(args)) {
        return;
      ***REMOVED***
      args.filter((arg) => arg.type === 'ObjectExpression').forEach((object) => validatePropNaming(node, object.properties));
    ***REMOVED***

    function getComponentTypeAnnotation(component) {
      // If this is a functional component that uses a global type, check it
      if (
        (component.node.type === 'FunctionDeclaration' || component.node.type === 'ArrowFunctionExpression')
        && component.node.params
        && component.node.params.length > 0
        && component.node.params[0].typeAnnotation
      ) {
        return component.node.params[0].typeAnnotation.typeAnnotation;
      ***REMOVED***

      if (
        !component.node.parent
        || component.node.parent.type !== 'VariableDeclarator'
        || !component.node.parent.id
        || component.node.parent.id.type !== 'Identifier'
        || !component.node.parent.id.typeAnnotation
        || !component.node.parent.id.typeAnnotation.typeAnnotation
      ) {
        return;
      ***REMOVED***

      const annotationTypeArguments = propsUtil.getTypeArguments(
        component.node.parent.id.typeAnnotation.typeAnnotation
      );
      if (
        annotationTypeArguments && (
          annotationTypeArguments.type === 'TSTypeParameterInstantiation'
          || annotationTypeArguments.type === 'TypeParameterInstantiation'
        )
      ) {
        return annotationTypeArguments.params.find(
          (param) => param.type === 'TSTypeReference' || param.type === 'GenericTypeAnnotation'
        );
      ***REMOVED***
    ***REMOVED***

    function findAllTypeAnnotations(identifier, node) {
      if (node.type === 'TSTypeLiteral' || node.type === 'ObjectTypeAnnotation' || node.type === 'TSInterfaceBody') {
        const currentNode = [].concat(
          objectTypeAnnotations.get(identifier.name) || [],
          node
        );
        objectTypeAnnotations.set(identifier.name, currentNode);
      ***REMOVED*** else if (
        node.type === 'TSParenthesizedType'
        && (
          node.typeAnnotation.type === 'TSIntersectionType'
          || node.typeAnnotation.type === 'TSUnionType'
        )
      ) {
        node.typeAnnotation.types.forEach((type) => {
          findAllTypeAnnotations(identifier, type);
        ***REMOVED***);
      ***REMOVED*** else if (
        node.type === 'TSIntersectionType'
        || node.type === 'TSUnionType'
        || node.type === 'IntersectionTypeAnnotation'
        || node.type === 'UnionTypeAnnotation'
      ) {
        node.types.forEach((type) => {
          findAllTypeAnnotations(identifier, type);
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***

    // --------------------------------------------------------------------------
    // Public
    // --------------------------------------------------------------------------

    return {
      'ClassProperty, PropertyDefinition'(node) {
        if (!rule || !propsUtil.isPropTypesDeclaration(node)) {
          return;
        ***REMOVED***
        if (
          node.value
          && astUtil.isCallExpression(node.value)
          && propWrapperUtil.isPropWrapperFunction(
            context,
            getText(context, node.value.callee)
          )
        ) {
          checkPropWrapperArguments(node, node.value.arguments);
        ***REMOVED***
        if (node.value && node.value.properties) {
          validatePropNaming(node, node.value.properties);
        ***REMOVED***
        if (node.typeAnnotation && node.typeAnnotation.typeAnnotation) {
          validatePropNaming(node, node.typeAnnotation.typeAnnotation.properties);
        ***REMOVED***
      ***REMOVED***,

      MemberExpression(node) {
        if (!rule || !propsUtil.isPropTypesDeclaration(node)) {
          return;
        ***REMOVED***
        const component = utils.getRelatedComponent(node);
        if (!component || !node.parent.right) {
          return;
        ***REMOVED***
        const right = node.parent.right;
        if (
          astUtil.isCallExpression(right)
          && propWrapperUtil.isPropWrapperFunction(
            context,
            getText(context, right.callee)
          )
        ) {
          checkPropWrapperArguments(component.node, right.arguments);
          return;
        ***REMOVED***
        validatePropNaming(component.node, node.parent.right.properties);
      ***REMOVED***,

      ObjectExpression(node) {
        if (!rule) {
          return;
        ***REMOVED***

        // Search for the proptypes declaration
        node.properties.forEach((property) => {
          if (!propsUtil.isPropTypesDeclaration(property)) {
            return;
          ***REMOVED***
          validatePropNaming(node, property.value.properties);
        ***REMOVED***);
      ***REMOVED***,

      TypeAlias(node) {
        findAllTypeAnnotations(node.id, node.right);
      ***REMOVED***,

      TSTypeAliasDeclaration(node) {
        findAllTypeAnnotations(node.id, node.typeAnnotation);
      ***REMOVED***,

      TSInterfaceDeclaration(node) {
        findAllTypeAnnotations(node.id, node.body);
      ***REMOVED***,

      // eslint-disable-next-line object-shorthand
      'Program:exit'() {
        if (!rule) {
          return;
        ***REMOVED***

        values(components.list()).forEach((component) => {
          const annotation = getComponentTypeAnnotation(component);

          if (annotation) {
            let propType;
            if (annotation.type === 'GenericTypeAnnotation') {
              propType = objectTypeAnnotations.get(annotation.id.name);
            ***REMOVED*** else if (annotation.type === 'ObjectTypeAnnotation' || annotation.type === 'TSTypeLiteral') {
              propType = annotation;
            ***REMOVED*** else if (annotation.type === 'TSTypeReference') {
              propType = objectTypeAnnotations.get(annotation.typeName.name);
            ***REMOVED*** else if (annotation.type === 'TSIntersectionType') {
              propType = flatMap(annotation.types, (type) => (
                type.type === 'TSTypeReference'
                  ? objectTypeAnnotations.get(type.typeName.name)
                  : type
              ));
            ***REMOVED***

            if (propType) {
              [].concat(propType).filter(Boolean).forEach((prop) => {
                validatePropNaming(
                  component.node,
                  prop.properties || prop.members || prop.body
                );
              ***REMOVED***);
            ***REMOVED***
          ***REMOVED***

          if (component.invalidProps && component.invalidProps.length > 0) {
            reportInvalidNaming(component);
          ***REMOVED***
        ***REMOVED***);

        // Reset cache
        objectTypeAnnotations.clear();
      ***REMOVED***,
    ***REMOVED***;
  ***REMOVED***),
***REMOVED***;
