// eslint-plugin-react/lib/rules/checked-requires-onchange-or-readonly.js
/**
 * @fileoverview Enforce the use of the 'onChange' or 'readonly' attribute when 'checked' is used'
 * @author Jaesoekjjang
 */

'use strict';

const ASTUtils = require('jsx-ast-utils');
const flatMap = require('array.prototype.flatmap');
const isCreateElement = require('../util/isCreateElement');
const report = require('../util/report');
const docsUrl = require('../util/docsUrl');

const messages = {
  missingProperty: '`checked` should be used with either `onChange` or `readOnly`.',
  exclusiveCheckedAttribute: 'Use either `checked` or `defaultChecked`, but not both.',
***REMOVED***;

const targetPropSet = new Set(['checked', 'onChange', 'readOnly', 'defaultChecked']);

const defaultOptions = {
  ignoreMissingProperties: false,
  ignoreExclusiveCheckedAttribute: false,
***REMOVED***;

/**
 * @param {object[]***REMOVED*** properties
 * @param {string***REMOVED*** keyName
 * @returns {Set<string>***REMOVED***
 */
function extractTargetProps(properties, keyName) {
  return new Set(
    flatMap(
      properties,
      (prop) => (
        prop[keyName] && targetPropSet.has(prop[keyName].name)
          ? [prop[keyName].name]
          : []
      )
    )
  );
***REMOVED***

/** @type {import('eslint').Rule.RuleModule***REMOVED*** */
module.exports = {
  meta: {
    docs: {
      description: 'Enforce using `onChange` or `readonly` attribute when `checked` is used',
      category: 'Best Practices',
      recommended: false,
      url: docsUrl('checked-requires-onchange-or-readonly'),
    ***REMOVED***,
    messages,
    schema: [{
      additionalProperties: false,
      properties: {
        ignoreMissingProperties: {
          type: 'boolean',
        ***REMOVED***,
        ignoreExclusiveCheckedAttribute: {
          type: 'boolean',
        ***REMOVED***,
      ***REMOVED***,
    ***REMOVED***],
  ***REMOVED***,
  create(context) {
    const options = Object.assign({***REMOVED***, defaultOptions, context.options[0]);

    function reportMissingProperty(node) {
      report(
        context,
        messages.missingProperty,
        'missingProperty',
        { node ***REMOVED***
      );
    ***REMOVED***

    function reportExclusiveCheckedAttribute(node) {
      report(
        context,
        messages.exclusiveCheckedAttribute,
        'exclusiveCheckedAttribute',
        { node ***REMOVED***
      );
    ***REMOVED***

    /**
     * @param {ASTNode***REMOVED*** node
     * @param {Set<string>***REMOVED*** propSet
     * @returns {void***REMOVED***
     */
    const checkAttributesAndReport = (node, propSet) => {
      if (!propSet.has('checked')) {
        return;
      ***REMOVED***

      if (!options.ignoreExclusiveCheckedAttribute && propSet.has('defaultChecked')) {
        reportExclusiveCheckedAttribute(node);
      ***REMOVED***

      if (
        !options.ignoreMissingProperties
        && !(propSet.has('onChange') || propSet.has('readOnly'))
      ) {
        reportMissingProperty(node);
      ***REMOVED***
    ***REMOVED***;

    return {
      JSXOpeningElement(node) {
        if (ASTUtils.elementType(node) !== 'input') {
          return;
        ***REMOVED***

        const propSet = extractTargetProps(node.attributes, 'name');
        checkAttributesAndReport(node, propSet);
      ***REMOVED***,
      CallExpression(node) {
        if (!isCreateElement(context, node)) {
          return;
        ***REMOVED***

        const firstArg = node.arguments[0];
        const secondArg = node.arguments[1];
        if (
          !firstArg
          || firstArg.type !== 'Literal'
          || firstArg.value !== 'input'
        ) {
          return;
        ***REMOVED***

        if (!secondArg || secondArg.type !== 'ObjectExpression') {
          return;
        ***REMOVED***

        const propSet = extractTargetProps(secondArg.properties, 'key');
        checkAttributesAndReport(node, propSet);
      ***REMOVED***,
    ***REMOVED***;
  ***REMOVED***,
***REMOVED***;
