// eslint-plugin-react/lib/rules/forbid-dom-props.js
/**
 * @fileoverview Forbid certain props on DOM Nodes
 * @author David Vázquez
 */

'use strict';

const docsUrl = require('../util/docsUrl');
const report = require('../util/report');

// ------------------------------------------------------------------------------
// Constants
// ------------------------------------------------------------------------------

const DEFAULTS = [];

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

/**
 * @param {Map<string, object>***REMOVED*** forbidMap // { disallowList: null | string[], message: null | string ***REMOVED***
 * @param {string***REMOVED*** prop
 * @param {string***REMOVED*** tagName
 * @returns {boolean***REMOVED***
 */
function isForbidden(forbidMap, prop, tagName) {
  const options = forbidMap.get(prop);
  return options && (
    typeof tagName === 'undefined'
    || !options.disallowList
    || options.disallowList.indexOf(tagName) !== -1
  );
***REMOVED***

const messages = {
  propIsForbidden: 'Prop "{{prop***REMOVED******REMOVED***" is forbidden on DOM Nodes',
***REMOVED***;

/** @type {import('eslint').Rule.RuleModule***REMOVED*** */
module.exports = {
  meta: {
    docs: {
      description: 'Disallow certain props on DOM Nodes',
      category: 'Best Practices',
      recommended: false,
      url: docsUrl('forbid-dom-props'),
    ***REMOVED***,

    messages,

    schema: [{
      type: 'object',
      properties: {
        forbid: {
          type: 'array',
          items: {
            anyOf: [{
              type: 'string',
            ***REMOVED***, {
              type: 'object',
              properties: {
                propName: {
                  type: 'string',
                ***REMOVED***,
                disallowedFor: {
                  type: 'array',
                  uniqueItems: true,
                  items: {
                    type: 'string',
                  ***REMOVED***,
                ***REMOVED***,
                message: {
                  type: 'string',
                ***REMOVED***,
              ***REMOVED***,
            ***REMOVED***],
            minLength: 1,
          ***REMOVED***,
          uniqueItems: true,
        ***REMOVED***,
      ***REMOVED***,
      additionalProperties: false,
    ***REMOVED***],
  ***REMOVED***,

  create(context) {
    const configuration = context.options[0] || {***REMOVED***;
    const forbid = new Map((configuration.forbid || DEFAULTS).map((value) => {
      const propName = typeof value === 'string' ? value : value.propName;
      return [propName, {
        disallowList: typeof value === 'string' ? null : (value.disallowedFor || null),
        message: typeof value === 'string' ? null : value.message,
      ***REMOVED***];
    ***REMOVED***));

    return {
      JSXAttribute(node) {
        const tag = node.parent.name.name;
        if (!(tag && typeof tag === 'string' && tag[0] !== tag[0].toUpperCase())) {
          // This is a Component, not a DOM node, so exit.
          return;
        ***REMOVED***

        const prop = node.name.name;

        if (!isForbidden(forbid, prop, tag)) {
          return;
        ***REMOVED***

        const customMessage = forbid.get(prop).message;

        report(context, customMessage || messages.propIsForbidden, !customMessage && 'propIsForbidden', {
          node,
          data: {
            prop,
          ***REMOVED***,
        ***REMOVED***);
      ***REMOVED***,
    ***REMOVED***;
  ***REMOVED***,
***REMOVED***;
