// eslint-plugin-react/lib/rules/forbid-prop-types.js
/**
 * @fileoverview Forbid certain propTypes
 */

'use strict';

const variableUtil = require('../util/variable');
const propsUtil = require('../util/props');
const astUtil = require('../util/ast');
const docsUrl = require('../util/docsUrl');
const propWrapperUtil = require('../util/propWrapper');
const report = require('../util/report');
const getText = require('../util/eslint').getText;

// ------------------------------------------------------------------------------
// Constants
// ------------------------------------------------------------------------------

const DEFAULTS = ['any', 'array', 'object'];

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

const messages = {
  forbiddenPropType: 'Prop type "{{target***REMOVED******REMOVED***" is forbidden',
***REMOVED***;

/** @type {import('eslint').Rule.RuleModule***REMOVED*** */
module.exports = {
  meta: {
    docs: {
      description: 'Disallow certain propTypes',
      category: 'Best Practices',
      recommended: false,
      url: docsUrl('forbid-prop-types'),
    ***REMOVED***,

    messages,

    schema: [{
      type: 'object',
      properties: {
        forbid: {
          type: 'array',
          items: {
            type: 'string',
          ***REMOVED***,
        ***REMOVED***,
        checkContextTypes: {
          type: 'boolean',
        ***REMOVED***,
        checkChildContextTypes: {
          type: 'boolean',
        ***REMOVED***,
      ***REMOVED***,
      additionalProperties: true,
    ***REMOVED***],
  ***REMOVED***,

  create(context) {
    const configuration = context.options[0] || {***REMOVED***;
    const checkContextTypes = configuration.checkContextTypes || false;
    const checkChildContextTypes = configuration.checkChildContextTypes || false;
    let propTypesPackageName = null;
    let reactPackageName = null;
    let isForeignPropTypesPackage = false;

    function isPropTypesPackage(node) {
      return (
        node.type === 'Identifier'
        && (
          node.name === null
          || node.name === propTypesPackageName
          || !isForeignPropTypesPackage
        )
      ) || (
        node.type === 'MemberExpression'
        && (
          node.object.name === null
          || node.object.name === reactPackageName
          || !isForeignPropTypesPackage
        )
      );
    ***REMOVED***

    function isForbidden(type) {
      const forbid = configuration.forbid || DEFAULTS;
      return forbid.indexOf(type) >= 0;
    ***REMOVED***

    function reportIfForbidden(type, declaration, target) {
      if (isForbidden(type)) {
        report(context, messages.forbiddenPropType, 'forbiddenPropType', {
          node: declaration,
          data: {
            target,
          ***REMOVED***,
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***

    function shouldCheckContextTypes(node) {
      if (checkContextTypes && propsUtil.isContextTypesDeclaration(node)) {
        return true;
      ***REMOVED***
      return false;
    ***REMOVED***

    function shouldCheckChildContextTypes(node) {
      if (checkChildContextTypes && propsUtil.isChildContextTypesDeclaration(node)) {
        return true;
      ***REMOVED***
      return false;
    ***REMOVED***

    /**
     * Checks if propTypes declarations are forbidden
     * @param {Array***REMOVED*** declarations The array of AST nodes being checked.
     * @returns {void***REMOVED***
     */
    function checkProperties(declarations) {
      if (declarations) {
        declarations.forEach((declaration) => {
          if (declaration.type !== 'Property') {
            return;
          ***REMOVED***
          let target;
          let value = declaration.value;
          if (
            value.type === 'MemberExpression'
            && value.property
            && value.property.name
            && value.property.name === 'isRequired'
          ) {
            value = value.object;
          ***REMOVED***
          if (astUtil.isCallExpression(value)) {
            if (!isPropTypesPackage(value.callee)) {
              return;
            ***REMOVED***
            value.arguments.forEach((arg) => {
              const name = arg.type === 'MemberExpression' ? arg.property.name : arg.name;
              reportIfForbidden(name, declaration, name);
            ***REMOVED***);
            value = value.callee;
          ***REMOVED***
          if (!isPropTypesPackage(value)) {
            return;
          ***REMOVED***
          if (value.property) {
            target = value.property.name;
          ***REMOVED*** else if (value.type === 'Identifier') {
            target = value.name;
          ***REMOVED***
          reportIfForbidden(target, declaration, target);
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***

    function checkNode(node) {
      if (!node) {
        return;
      ***REMOVED***

      if (node.type === 'ObjectExpression') {
        checkProperties(node.properties);
      ***REMOVED*** else if (node.type === 'Identifier') {
        const propTypesObject = variableUtil.findVariableByName(context, node, node.name);
        if (propTypesObject && propTypesObject.properties) {
          checkProperties(propTypesObject.properties);
        ***REMOVED***
      ***REMOVED*** else if (astUtil.isCallExpression(node)) {
        const innerNode = node.arguments && node.arguments[0];
        if (
          propWrapperUtil.isPropWrapperFunction(context, getText(context, node.callee))
            && innerNode
        ) {
          checkNode(innerNode);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    return {
      ImportDeclaration(node) {
        if (node.source && node.source.value === 'prop-types') { // import PropType from "prop-types"
          if (node.specifiers.length > 0) {
            propTypesPackageName = node.specifiers[0].local.name;
          ***REMOVED***
        ***REMOVED*** else if (node.source && node.source.value === 'react') { // import { PropTypes ***REMOVED*** from "react"
          if (node.specifiers.length > 0) {
            reactPackageName = node.specifiers[0].local.name; // guard against accidental anonymous `import "react"`
          ***REMOVED***
          if (node.specifiers.length >= 1) {
            const propTypesSpecifier = node.specifiers.find((specifier) => (
              'imported' in specifier
              && specifier.imported
              && 'name' in specifier.imported
              && specifier.imported.name === 'PropTypes'
            ));
            if (propTypesSpecifier) {
              propTypesPackageName = propTypesSpecifier.local.name;
            ***REMOVED***
          ***REMOVED***
        ***REMOVED*** else { // package is not imported from "react" or "prop-types"
          // eslint-disable-next-line no-lonely-if
          if (node.specifiers.some((x) => x.local.name === 'PropTypes')) { // assert: node.specifiers.length > 1
            isForeignPropTypesPackage = true;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***,

      'ClassProperty, PropertyDefinition'(node) {
        if (
          !propsUtil.isPropTypesDeclaration(node)
          && !isPropTypesPackage(node)
          && !shouldCheckContextTypes(node)
          && !shouldCheckChildContextTypes(node)
        ) {
          return;
        ***REMOVED***
        checkNode(node.value);
      ***REMOVED***,

      MemberExpression(node) {
        if (
          !propsUtil.isPropTypesDeclaration(node)
          && !isPropTypesPackage(node)
          && !shouldCheckContextTypes(node)
          && !shouldCheckChildContextTypes(node)
        ) {
          return;
        ***REMOVED***

        checkNode('right' in node.parent && node.parent.right);
      ***REMOVED***,

      CallExpression(node) {
        if (
          node.callee.type === 'MemberExpression'
          && node.callee.object
          && !isPropTypesPackage(node.callee.object)
          && !propsUtil.isPropTypesDeclaration(node.callee)
        ) {
          return;
        ***REMOVED***

        if (
          node.arguments.length > 0
          && (
            ('name' in node.callee && node.callee.name === 'shape')
            || astUtil.getPropertyName(node.callee) === 'shape'
          )
        ) {
          checkProperties('properties' in node.arguments[0] && node.arguments[0].properties);
        ***REMOVED***
      ***REMOVED***,

      MethodDefinition(node) {
        if (
          !propsUtil.isPropTypesDeclaration(node)
          && !isPropTypesPackage(node)
          && !shouldCheckContextTypes(node)
          && !shouldCheckChildContextTypes(node)
        ) {
          return;
        ***REMOVED***

        const returnStatement = astUtil.findReturnStatement(node);

        if (returnStatement && returnStatement.argument) {
          checkNode(returnStatement.argument);
        ***REMOVED***
      ***REMOVED***,

      ObjectExpression(node) {
        node.properties.forEach((property) => {
          if (!('key' in property) || !property.key) {
            return;
          ***REMOVED***

          if (
            !propsUtil.isPropTypesDeclaration(property)
            && !isPropTypesPackage(property)
            && !shouldCheckContextTypes(property)
            && !shouldCheckChildContextTypes(property)
          ) {
            return;
          ***REMOVED***
          if (property.value.type === 'ObjectExpression') {
            checkProperties(property.value.properties);
          ***REMOVED***
        ***REMOVED***);
      ***REMOVED***,

    ***REMOVED***;
  ***REMOVED***,
***REMOVED***;
