// eslint-plugin-react/lib/rules/forbid-component-props.js
/**
 * @fileoverview Forbid certain props on components
 * @author Joe Lencioni
 */

'use strict';

const minimatch = require('minimatch');
const docsUrl = require('../util/docsUrl');
const report = require('../util/report');

// ------------------------------------------------------------------------------
// Constants
// ------------------------------------------------------------------------------

const DEFAULTS = ['className', 'style'];

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

const messages = {
  propIsForbidden: 'Prop "{{prop***REMOVED******REMOVED***" is forbidden on Components',
***REMOVED***;

/** @type {import('eslint').Rule.RuleModule***REMOVED*** */
module.exports = {
  meta: {
    docs: {
      description: 'Disallow certain props on components',
      category: 'Best Practices',
      recommended: false,
      url: docsUrl('forbid-component-props'),
    ***REMOVED***,

    messages,

    schema: [{
      type: 'object',
      properties: {
        forbid: {
          type: 'array',
          items: {
            anyOf: [
              { type: 'string' ***REMOVED***,
              {
                type: 'object',
                properties: {
                  propName: { type: 'string' ***REMOVED***,
                  allowedFor: {
                    type: 'array',
                    uniqueItems: true,
                    items: { type: 'string' ***REMOVED***,
                  ***REMOVED***,
                  allowedForPatterns: {
                    type: 'array',
                    uniqueItems: true,
                    items: { type: 'string' ***REMOVED***,
                  ***REMOVED***,
                  message: { type: 'string' ***REMOVED***,
                ***REMOVED***,
                additionalProperties: false,
              ***REMOVED***,
              {
                type: 'object',
                properties: {
                  propName: { type: 'string' ***REMOVED***,
                  disallowedFor: {
                    type: 'array',
                    uniqueItems: true,
                    minItems: 1,
                    items: { type: 'string' ***REMOVED***,
                  ***REMOVED***,
                  disallowedForPatterns: {
                    type: 'array',
                    uniqueItems: true,
                    minItems: 1,
                    items: { type: 'string' ***REMOVED***,
                  ***REMOVED***,
                  message: { type: 'string' ***REMOVED***,
                ***REMOVED***,
                anyOf: [
                  { required: ['disallowedFor'] ***REMOVED***,
                  { required: ['disallowedForPatterns'] ***REMOVED***,
                ],
                additionalProperties: false,
              ***REMOVED***,
              {
                type: 'object',
                properties: {
                  propNamePattern: { type: 'string' ***REMOVED***,
                  allowedFor: {
                    type: 'array',
                    uniqueItems: true,
                    items: { type: 'string' ***REMOVED***,
                  ***REMOVED***,
                  allowedForPatterns: {
                    type: 'array',
                    uniqueItems: true,
                    items: { type: 'string' ***REMOVED***,
                  ***REMOVED***,
                  message: { type: 'string' ***REMOVED***,
                ***REMOVED***,
                additionalProperties: false,
              ***REMOVED***,
              {
                type: 'object',
                properties: {
                  propNamePattern: { type: 'string' ***REMOVED***,
                  disallowedFor: {
                    type: 'array',
                    uniqueItems: true,
                    minItems: 1,
                    items: { type: 'string' ***REMOVED***,
                  ***REMOVED***,
                  disallowedForPatterns: {
                    type: 'array',
                    uniqueItems: true,
                    minItems: 1,
                    items: { type: 'string' ***REMOVED***,
                  ***REMOVED***,
                  message: { type: 'string' ***REMOVED***,
                ***REMOVED***,
                anyOf: [
                  { required: ['disallowedFor'] ***REMOVED***,
                  { required: ['disallowedForPatterns'] ***REMOVED***,
                ],
                additionalProperties: false,
              ***REMOVED***,
            ],
          ***REMOVED***,
        ***REMOVED***,
      ***REMOVED***,
    ***REMOVED***],
  ***REMOVED***,

  create(context) {
    const configuration = context.options[0] || {***REMOVED***;
    const forbid = new Map((configuration.forbid || DEFAULTS).map((value) => {
      const propName = typeof value === 'string' ? value : value.propName;
      const propPattern = value.propNamePattern;
      const prop = propName || propPattern;
      const options = {
        allowList: [].concat(value.allowedFor || []),
        allowPatternList: [].concat(value.allowedForPatterns || []),
        disallowList: [].concat(value.disallowedFor || []),
        disallowPatternList: [].concat(value.disallowedForPatterns || []),
        message: typeof value === 'string' ? null : value.message,
        isPattern: !!value.propNamePattern,
      ***REMOVED***;
      return [prop, options];
    ***REMOVED***));

    function getPropOptions(prop) {
      // Get config options having pattern
      const propNamePatternArray = Array.from(forbid.entries()).filter((propEntry) => propEntry[1].isPattern);
      // Match current prop with pattern options, return if matched
      const propNamePattern = propNamePatternArray.find((propPatternVal) => minimatch(prop, propPatternVal[0]));
      // Get options for matched propNamePattern
      const propNamePatternOptions = propNamePattern && propNamePattern[1];

      const options = forbid.get(prop) || propNamePatternOptions;
      return options;
    ***REMOVED***

    function isForbidden(prop, tagName) {
      const options = getPropOptions(prop);
      if (!options) {
        return false;
      ***REMOVED***

      function checkIsTagForbiddenByAllowOptions() {
        if (options.allowList.indexOf(tagName) !== -1) {
          return false;
        ***REMOVED***

        if (options.allowPatternList.length === 0) {
          return true;
        ***REMOVED***

        return options.allowPatternList.every(
          (pattern) => !minimatch(tagName, pattern)
        );
      ***REMOVED***

      function checkIsTagForbiddenByDisallowOptions() {
        if (options.disallowList.indexOf(tagName) !== -1) {
          return true;
        ***REMOVED***

        if (options.disallowPatternList.length === 0) {
          return false;
        ***REMOVED***

        return options.disallowPatternList.some(
          (pattern) => minimatch(tagName, pattern)
        );
      ***REMOVED***

      const hasDisallowOptions = options.disallowList.length > 0 || options.disallowPatternList.length > 0;

      // disallowList should have a least one item (schema configuration)
      const isTagForbidden = hasDisallowOptions
        ? checkIsTagForbiddenByDisallowOptions()
        : checkIsTagForbiddenByAllowOptions();

      // if the tagName is undefined (`<this.something>`), we assume it's a forbidden element
      return typeof tagName === 'undefined' || isTagForbidden;
    ***REMOVED***

    return {
      JSXAttribute(node) {
        const parentName = node.parent.name;
        // Extract a component name when using a "namespace", e.g. `<AntdLayout.Content />`.
        const tag = parentName.name || `${parentName.object.name***REMOVED***.${parentName.property.name***REMOVED***`;
        const componentName = parentName.name || parentName.property.name;
        if (componentName && typeof componentName[0] === 'string' && componentName[0] !== componentName[0].toUpperCase()) {
          // This is a DOM node, not a Component, so exit.
          return;
        ***REMOVED***

        const prop = node.name.name;

        if (!isForbidden(prop, tag)) {
          return;
        ***REMOVED***

        const customMessage = getPropOptions(prop).message;

        report(context, customMessage || messages.propIsForbidden, !customMessage && 'propIsForbidden', {
          node,
          data: {
            prop,
          ***REMOVED***,
        ***REMOVED***);
      ***REMOVED***,
    ***REMOVED***;
  ***REMOVED***,
***REMOVED***;
