// eslint-plugin-react/lib/rules/jsx-curly-newline.js
/**
 * @fileoverview enforce consistent line breaks inside jsx curly
 */

'use strict';

const docsUrl = require('../util/docsUrl');
const eslintUtil = require('../util/eslint');
const report = require('../util/report');

const getSourceCode = eslintUtil.getSourceCode;
const getText = eslintUtil.getText;

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

function getNormalizedOption(context) {
  const rawOption = context.options[0] || 'consistent';

  if (rawOption === 'consistent') {
    return {
      multiline: 'consistent',
      singleline: 'consistent',
    ***REMOVED***;
  ***REMOVED***

  if (rawOption === 'never') {
    return {
      multiline: 'forbid',
      singleline: 'forbid',
    ***REMOVED***;
  ***REMOVED***

  return {
    multiline: rawOption.multiline || 'consistent',
    singleline: rawOption.singleline || 'consistent',
  ***REMOVED***;
***REMOVED***

const messages = {
  expectedBefore: 'Expected newline before \'***REMOVED***\'.',
  expectedAfter: 'Expected newline after \'{\'.',
  unexpectedBefore: 'Unexpected newline before \'***REMOVED***\'.',
  unexpectedAfter: 'Unexpected newline after \'{\'.',
***REMOVED***;

/** @type {import('eslint').Rule.RuleModule***REMOVED*** */
module.exports = {
  meta: {
    type: 'layout',

    docs: {
      description: 'Enforce consistent linebreaks in curly braces in JSX attributes and expressions',
      category: 'Stylistic Issues',
      recommended: false,
      url: docsUrl('jsx-curly-newline'),
    ***REMOVED***,

    fixable: 'whitespace',

    schema: [
      {
        anyOf: [
          {
            enum: ['consistent', 'never'],
          ***REMOVED***,
          {
            type: 'object',
            properties: {
              singleline: { enum: ['consistent', 'require', 'forbid'] ***REMOVED***,
              multiline: { enum: ['consistent', 'require', 'forbid'] ***REMOVED***,
            ***REMOVED***,
            additionalProperties: false,
          ***REMOVED***,
        ],
      ***REMOVED***,
    ],

    messages,
  ***REMOVED***,

  create(context) {
    const sourceCode = getSourceCode(context);
    const option = getNormalizedOption(context);

    // ----------------------------------------------------------------------
    // Helpers
    // ----------------------------------------------------------------------

    /**
     * Determines whether two adjacent tokens are on the same line.
     * @param {Object***REMOVED*** left - The left token object.
     * @param {Object***REMOVED*** right - The right token object.
     * @returns {boolean***REMOVED*** Whether or not the tokens are on the same line.
     */
    function isTokenOnSameLine(left, right) {
      return left.loc.end.line === right.loc.start.line;
    ***REMOVED***

    /**
     * Determines whether there should be newlines inside curlys
     * @param {ASTNode***REMOVED*** expression The expression contained in the curlys
     * @param {boolean***REMOVED*** hasLeftNewline `true` if the left curly has a newline in the current code.
     * @returns {boolean***REMOVED*** `true` if there should be newlines inside the function curlys
     */
    function shouldHaveNewlines(expression, hasLeftNewline) {
      const isMultiline = expression.loc.start.line !== expression.loc.end.line;

      switch (isMultiline ? option.multiline : option.singleline) {
        case 'forbid': return false;
        case 'require': return true;
        case 'consistent':
        default: return hasLeftNewline;
      ***REMOVED***
    ***REMOVED***

    /**
     * Validates curlys
     * @param {Object***REMOVED*** curlys An object with keys `leftParen` for the left paren token, and `rightParen` for the right paren token
     * @param {ASTNode***REMOVED*** expression The expression inside the curly
     * @returns {void***REMOVED***
     */
    function validateCurlys(curlys, expression) {
      const leftCurly = curlys.leftCurly;
      const rightCurly = curlys.rightCurly;
      const tokenAfterLeftCurly = sourceCode.getTokenAfter(leftCurly);
      const tokenBeforeRightCurly = sourceCode.getTokenBefore(rightCurly);
      const hasLeftNewline = !isTokenOnSameLine(leftCurly, tokenAfterLeftCurly);
      const hasRightNewline = !isTokenOnSameLine(tokenBeforeRightCurly, rightCurly);
      const needsNewlines = shouldHaveNewlines(expression, hasLeftNewline);

      if (hasLeftNewline && !needsNewlines) {
        report(context, messages.unexpectedAfter, 'unexpectedAfter', {
          node: leftCurly,
          fix(fixer) {
            return getText(context)
              .slice(leftCurly.range[1], tokenAfterLeftCurly.range[0])
              .trim()
              ? null // If there is a comment between the { and the first element, don't do a fix.
              : fixer.removeRange([leftCurly.range[1], tokenAfterLeftCurly.range[0]]);
          ***REMOVED***,
        ***REMOVED***);
      ***REMOVED*** else if (!hasLeftNewline && needsNewlines) {
        report(context, messages.expectedAfter, 'expectedAfter', {
          node: leftCurly,
          fix: (fixer) => fixer.insertTextAfter(leftCurly, '\n'),
        ***REMOVED***);
      ***REMOVED***

      if (hasRightNewline && !needsNewlines) {
        report(context, messages.unexpectedBefore, 'unexpectedBefore', {
          node: rightCurly,
          fix(fixer) {
            return getText(context)
              .slice(tokenBeforeRightCurly.range[1], rightCurly.range[0])
              .trim()
              ? null // If there is a comment between the last element and the ***REMOVED***, don't do a fix.
              : fixer.removeRange([
                tokenBeforeRightCurly.range[1],
                rightCurly.range[0],
              ]);
          ***REMOVED***,
        ***REMOVED***);
      ***REMOVED*** else if (!hasRightNewline && needsNewlines) {
        report(context, messages.expectedBefore, 'expectedBefore', {
          node: rightCurly,
          fix: (fixer) => fixer.insertTextBefore(rightCurly, '\n'),
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***

    // ----------------------------------------------------------------------
    // Public
    // ----------------------------------------------------------------------

    return {
      JSXExpressionContainer(node) {
        const curlyTokens = {
          leftCurly: sourceCode.getFirstToken(node),
          rightCurly: sourceCode.getLastToken(node),
        ***REMOVED***;
        validateCurlys(curlyTokens, node.expression);
      ***REMOVED***,
    ***REMOVED***;
  ***REMOVED***,
***REMOVED***;
