// eslint-plugin-react/lib/rules/jsx-curly-spacing.js
/**
 * @fileoverview Enforce or disallow spaces inside of curly braces in JSX attributes.
 * @author Jamund Ferguson
 * @author Brandyn Bennett
 * @author Michael Ficarra
 * @author Vignesh Anand
 * @author Jamund Ferguson
 * @author Yannick Croissant
 * @author Erik Wendel
 */

'use strict';

const has = require('hasown');
const docsUrl = require('../util/docsUrl');
const getSourceCode = require('../util/eslint').getSourceCode;
const report = require('../util/report');

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

const SPACING = {
  always: 'always',
  never: 'never',
***REMOVED***;
const SPACING_VALUES = [SPACING.always, SPACING.never];

const messages = {
  noNewlineAfter: 'There should be no newline after \'{{token***REMOVED******REMOVED***\'',
  noNewlineBefore: 'There should be no newline before \'{{token***REMOVED******REMOVED***\'',
  noSpaceAfter: 'There should be no space after \'{{token***REMOVED******REMOVED***\'',
  noSpaceBefore: 'There should be no space before \'{{token***REMOVED******REMOVED***\'',
  spaceNeededAfter: 'A space is required after \'{{token***REMOVED******REMOVED***\'',
  spaceNeededBefore: 'A space is required before \'{{token***REMOVED******REMOVED***\'',
***REMOVED***;

/** @type {import('eslint').Rule.RuleModule***REMOVED*** */
module.exports = {
  meta: {
    docs: {
      description: 'Enforce or disallow spaces inside of curly braces in JSX attributes and expressions',
      category: 'Stylistic Issues',
      recommended: false,
      url: docsUrl('jsx-curly-spacing'),
    ***REMOVED***,
    fixable: 'code',

    messages,

    schema: {
      definitions: {
        basicConfig: {
          type: 'object',
          properties: {
            when: {
              enum: SPACING_VALUES,
            ***REMOVED***,
            allowMultiline: {
              type: 'boolean',
            ***REMOVED***,
            spacing: {
              type: 'object',
              properties: {
                objectLiterals: {
                  enum: SPACING_VALUES,
                ***REMOVED***,
              ***REMOVED***,
            ***REMOVED***,
          ***REMOVED***,
        ***REMOVED***,
        basicConfigOrBoolean: {
          anyOf: [{
            $ref: '#/definitions/basicConfig',
          ***REMOVED***, {
            type: 'boolean',
          ***REMOVED***],
        ***REMOVED***,
      ***REMOVED***,
      type: 'array',
      items: [{
        anyOf: [{
          allOf: [{
            $ref: '#/definitions/basicConfig',
          ***REMOVED***, {
            type: 'object',
            properties: {
              attributes: {
                $ref: '#/definitions/basicConfigOrBoolean',
              ***REMOVED***,
              children: {
                $ref: '#/definitions/basicConfigOrBoolean',
              ***REMOVED***,
            ***REMOVED***,
          ***REMOVED***],
        ***REMOVED***, {
          enum: SPACING_VALUES,
        ***REMOVED***],
      ***REMOVED***, {
        type: 'object',
        properties: {
          allowMultiline: {
            type: 'boolean',
          ***REMOVED***,
          spacing: {
            type: 'object',
            properties: {
              objectLiterals: {
                enum: SPACING_VALUES,
              ***REMOVED***,
            ***REMOVED***,
          ***REMOVED***,
        ***REMOVED***,
        additionalProperties: false,
      ***REMOVED***],
    ***REMOVED***,
  ***REMOVED***,

  create(context) {
    function normalizeConfig(configOrTrue, defaults, lastPass) {
      const config = configOrTrue === true ? {***REMOVED*** : configOrTrue;
      const when = config.when || defaults.when;
      const allowMultiline = has(config, 'allowMultiline') ? config.allowMultiline : defaults.allowMultiline;
      const spacing = config.spacing || {***REMOVED***;
      let objectLiteralSpaces = spacing.objectLiterals || defaults.objectLiteralSpaces;
      if (lastPass) {
        // On the final pass assign the values that should be derived from others if they are still undefined
        objectLiteralSpaces = objectLiteralSpaces || when;
      ***REMOVED***

      return {
        when,
        allowMultiline,
        objectLiteralSpaces,
      ***REMOVED***;
    ***REMOVED***

    const DEFAULT_WHEN = SPACING.never;
    const DEFAULT_ALLOW_MULTILINE = true;
    const DEFAULT_ATTRIBUTES = true;
    const DEFAULT_CHILDREN = false;

    let originalConfig = context.options[0] || {***REMOVED***;
    if (SPACING_VALUES.indexOf(originalConfig) !== -1) {
      originalConfig = Object.assign({ when: context.options[0] ***REMOVED***, context.options[1]);
    ***REMOVED***
    const defaultConfig = normalizeConfig(originalConfig, {
      when: DEFAULT_WHEN,
      allowMultiline: DEFAULT_ALLOW_MULTILINE,
    ***REMOVED***);
    const attributes = has(originalConfig, 'attributes') ? originalConfig.attributes : DEFAULT_ATTRIBUTES;
    const attributesConfig = attributes ? normalizeConfig(attributes, defaultConfig, true) : null;
    const children = has(originalConfig, 'children') ? originalConfig.children : DEFAULT_CHILDREN;
    const childrenConfig = children ? normalizeConfig(children, defaultConfig, true) : null;

    // --------------------------------------------------------------------------
    // Helpers
    // --------------------------------------------------------------------------

    /**
     * Determines whether two adjacent tokens have a newline between them.
     * @param {Object***REMOVED*** left - The left token object.
     * @param {Object***REMOVED*** right - The right token object.
     * @returns {boolean***REMOVED*** Whether or not there is a newline between the tokens.
     */
    function isMultiline(left, right) {
      return left.loc.end.line !== right.loc.start.line;
    ***REMOVED***

    /**
     * Trims text of whitespace between two ranges
     * @param {Fixer***REMOVED*** fixer - the eslint fixer object
     * @param {number***REMOVED*** fromLoc - the start location
     * @param {number***REMOVED*** toLoc - the end location
     * @param {string***REMOVED*** mode - either 'start' or 'end'
     * @param {string=***REMOVED*** spacing - a spacing value that will optionally add a space to the removed text
     * @returns {Object|*|{range, text***REMOVED******REMOVED***
     */
    function fixByTrimmingWhitespace(fixer, fromLoc, toLoc, mode, spacing) {
      let replacementText = getSourceCode(context).text.slice(fromLoc, toLoc);
      if (mode === 'start') {
        replacementText = replacementText.replace(/^\s+/gm, '');
      ***REMOVED*** else {
        replacementText = replacementText.replace(/\s+$/gm, '');
      ***REMOVED***
      if (spacing === SPACING.always) {
        if (mode === 'start') {
          replacementText += ' ';
        ***REMOVED*** else {
          replacementText = ` ${replacementText***REMOVED***`;
        ***REMOVED***
      ***REMOVED***
      return fixer.replaceTextRange([fromLoc, toLoc], replacementText);
    ***REMOVED***

    /**
    * Reports that there shouldn't be a newline after the first token
    * @param {ASTNode***REMOVED*** node - The node to report in the event of an error.
    * @param {Token***REMOVED*** token - The token to use for the report.
    * @param {string***REMOVED*** spacing
    * @returns {void***REMOVED***
    */
    function reportNoBeginningNewline(node, token, spacing) {
      report(context, messages.noNewlineAfter, 'noNewlineAfter', {
        node,
        loc: token.loc.start,
        data: {
          token: token.value,
        ***REMOVED***,
        fix(fixer) {
          const nextToken = getSourceCode(context).getTokenAfter(token);
          return fixByTrimmingWhitespace(fixer, token.range[1], nextToken.range[0], 'start', spacing);
        ***REMOVED***,
      ***REMOVED***);
    ***REMOVED***

    /**
    * Reports that there shouldn't be a newline before the last token
    * @param {ASTNode***REMOVED*** node - The node to report in the event of an error.
    * @param {Token***REMOVED*** token - The token to use for the report.
    * @param {string***REMOVED*** spacing
    * @returns {void***REMOVED***
    */
    function reportNoEndingNewline(node, token, spacing) {
      report(context, messages.noNewlineBefore, 'noNewlineBefore', {
        node,
        loc: token.loc.start,
        data: {
          token: token.value,
        ***REMOVED***,
        fix(fixer) {
          const previousToken = getSourceCode(context).getTokenBefore(token);
          return fixByTrimmingWhitespace(fixer, previousToken.range[1], token.range[0], 'end', spacing);
        ***REMOVED***,
      ***REMOVED***);
    ***REMOVED***

    /**
    * Reports that there shouldn't be a space after the first token
    * @param {ASTNode***REMOVED*** node - The node to report in the event of an error.
    * @param {Token***REMOVED*** token - The token to use for the report.
    * @returns {void***REMOVED***
    */
    function reportNoBeginningSpace(node, token) {
      report(context, messages.noSpaceAfter, 'noSpaceAfter', {
        node,
        loc: token.loc.start,
        data: {
          token: token.value,
        ***REMOVED***,
        fix(fixer) {
          const sourceCode = getSourceCode(context);
          const nextToken = sourceCode.getTokenAfter(token);
          let nextComment;

          // eslint >=4.x
          if (sourceCode.getCommentsAfter) {
            nextComment = sourceCode.getCommentsAfter(token);
          // eslint 3.x
          ***REMOVED*** else {
            const potentialComment = sourceCode.getTokenAfter(token, { includeComments: true ***REMOVED***);
            nextComment = nextToken === potentialComment ? [] : [potentialComment];
          ***REMOVED***

          // Take comments into consideration to narrow the fix range to what is actually affected. (See #1414)
          if (nextComment.length > 0) {
            return fixByTrimmingWhitespace(fixer, token.range[1], Math.min(nextToken.range[0], nextComment[0].range[0]), 'start');
          ***REMOVED***

          return fixByTrimmingWhitespace(fixer, token.range[1], nextToken.range[0], 'start');
        ***REMOVED***,
      ***REMOVED***);
    ***REMOVED***

    /**
    * Reports that there shouldn't be a space before the last token
    * @param {ASTNode***REMOVED*** node - The node to report in the event of an error.
    * @param {Token***REMOVED*** token - The token to use for the report.
    * @returns {void***REMOVED***
    */
    function reportNoEndingSpace(node, token) {
      report(context, messages.noSpaceBefore, 'noSpaceBefore', {
        node,
        loc: token.loc.start,
        data: {
          token: token.value,
        ***REMOVED***,
        fix(fixer) {
          const sourceCode = getSourceCode(context);
          const previousToken = sourceCode.getTokenBefore(token);
          let previousComment;

          // eslint >=4.x
          if (sourceCode.getCommentsBefore) {
            previousComment = sourceCode.getCommentsBefore(token);
          // eslint 3.x
          ***REMOVED*** else {
            const potentialComment = sourceCode.getTokenBefore(token, { includeComments: true ***REMOVED***);
            previousComment = previousToken === potentialComment ? [] : [potentialComment];
          ***REMOVED***

          // Take comments into consideration to narrow the fix range to what is actually affected. (See #1414)
          if (previousComment.length > 0) {
            return fixByTrimmingWhitespace(fixer, Math.max(previousToken.range[1], previousComment[0].range[1]), token.range[0], 'end');
          ***REMOVED***

          return fixByTrimmingWhitespace(fixer, previousToken.range[1], token.range[0], 'end');
        ***REMOVED***,
      ***REMOVED***);
    ***REMOVED***

    /**
    * Reports that there should be a space after the first token
    * @param {ASTNode***REMOVED*** node - The node to report in the event of an error.
    * @param {Token***REMOVED*** token - The token to use for the report.
    * @returns {void***REMOVED***
    */
    function reportRequiredBeginningSpace(node, token) {
      report(context, messages.spaceNeededAfter, 'spaceNeededAfter', {
        node,
        loc: token.loc.start,
        data: {
          token: token.value,
        ***REMOVED***,
        fix(fixer) {
          return fixer.insertTextAfter(token, ' ');
        ***REMOVED***,
      ***REMOVED***);
    ***REMOVED***

    /**
    * Reports that there should be a space before the last token
    * @param {ASTNode***REMOVED*** node - The node to report in the event of an error.
    * @param {Token***REMOVED*** token - The token to use for the report.
    * @returns {void***REMOVED***
    */
    function reportRequiredEndingSpace(node, token) {
      report(context, messages.spaceNeededBefore, 'spaceNeededBefore', {
        node,
        loc: token.loc.start,
        data: {
          token: token.value,
        ***REMOVED***,
        fix(fixer) {
          return fixer.insertTextBefore(token, ' ');
        ***REMOVED***,
      ***REMOVED***);
    ***REMOVED***

    /**
     * Determines if spacing in curly braces is valid.
     * @param {ASTNode***REMOVED*** node The AST node to check.
     * @returns {void***REMOVED***
     */
    function validateBraceSpacing(node) {
      let config;
      switch (node.parent.type) {
        case 'JSXAttribute':
        case 'JSXOpeningElement':
          config = attributesConfig;
          break;

        case 'JSXElement':
        case 'JSXFragment':
          config = childrenConfig;
          break;

        default:
          return;
      ***REMOVED***
      if (config === null) {
        return;
      ***REMOVED***

      const sourceCode = getSourceCode(context);
      const first = sourceCode.getFirstToken(node);
      const last = sourceCode.getLastToken(node);
      let second = sourceCode.getTokenAfter(first, { includeComments: true ***REMOVED***);
      let penultimate = sourceCode.getTokenBefore(last, { includeComments: true ***REMOVED***);

      if (!second) {
        second = sourceCode.getTokenAfter(first);
        const leadingComments = sourceCode.getNodeByRangeIndex(second.range[0]).leadingComments;
        second = leadingComments ? leadingComments[0] : second;
      ***REMOVED***
      if (!penultimate) {
        penultimate = sourceCode.getTokenBefore(last);
        const trailingComments = sourceCode.getNodeByRangeIndex(penultimate.range[0]).trailingComments;
        penultimate = trailingComments ? trailingComments[trailingComments.length - 1] : penultimate;
      ***REMOVED***

      const isObjectLiteral = first.value === second.value;
      const spacing = isObjectLiteral ? config.objectLiteralSpaces : config.when;
      if (spacing === SPACING.always) {
        if (!sourceCode.isSpaceBetweenTokens(first, second)) {
          reportRequiredBeginningSpace(node, first);
        ***REMOVED*** else if (!config.allowMultiline && isMultiline(first, second)) {
          reportNoBeginningNewline(node, first, spacing);
        ***REMOVED***
        if (!sourceCode.isSpaceBetweenTokens(penultimate, last)) {
          reportRequiredEndingSpace(node, last);
        ***REMOVED*** else if (!config.allowMultiline && isMultiline(penultimate, last)) {
          reportNoEndingNewline(node, last, spacing);
        ***REMOVED***
      ***REMOVED*** else if (spacing === SPACING.never) {
        if (isMultiline(first, second)) {
          if (!config.allowMultiline) {
            reportNoBeginningNewline(node, first, spacing);
          ***REMOVED***
        ***REMOVED*** else if (sourceCode.isSpaceBetweenTokens(first, second)) {
          reportNoBeginningSpace(node, first);
        ***REMOVED***
        if (isMultiline(penultimate, last)) {
          if (!config.allowMultiline) {
            reportNoEndingNewline(node, last, spacing);
          ***REMOVED***
        ***REMOVED*** else if (sourceCode.isSpaceBetweenTokens(penultimate, last)) {
          reportNoEndingSpace(node, last);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    // --------------------------------------------------------------------------
    // Public
    // --------------------------------------------------------------------------

    return {
      JSXExpressionContainer: validateBraceSpacing,
      JSXSpreadAttribute: validateBraceSpacing,
    ***REMOVED***;
  ***REMOVED***,
***REMOVED***;
