// eslint-plugin-react/lib/rules/jsx-sort-default-props.js
/**
 * @fileoverview Enforce default props alphabetical sorting
 * @author Vladimir Kattsov
 * @deprecated
 */

'use strict';

const variableUtil = require('../util/variable');
const docsUrl = require('../util/docsUrl');
const report = require('../util/report');
const log = require('../util/log');
const eslintUtil = require('../util/eslint');

const getFirstTokens = eslintUtil.getFirstTokens;
const getText = eslintUtil.getText;

let isWarnedForDeprecation = false;

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

const messages = {
  propsNotSorted: 'Default prop types declarations should be sorted alphabetically',
***REMOVED***;

/** @type {import('eslint').Rule.RuleModule***REMOVED*** */
module.exports = {
  meta: {
    deprecated: true,
    replacedBy: ['sort-default-props'],
    docs: {
      description: 'Enforce defaultProps declarations alphabetical sorting',
      category: 'Stylistic Issues',
      recommended: false,
      url: docsUrl('jsx-sort-default-props'),
    ***REMOVED***,
    // fixable: 'code',

    messages,

    schema: [{
      type: 'object',
      properties: {
        ignoreCase: {
          type: 'boolean',
        ***REMOVED***,
      ***REMOVED***,
      additionalProperties: false,
    ***REMOVED***],
  ***REMOVED***,

  create(context) {
    const configuration = context.options[0] || {***REMOVED***;
    const ignoreCase = configuration.ignoreCase || false;

    /**
     * Get properties name
     * @param {Object***REMOVED*** node - Property.
     * @returns {string***REMOVED*** Property name.
     */
    function getPropertyName(node) {
      if (node.key || ['MethodDefinition', 'Property'].indexOf(node.type) !== -1) {
        return node.key.name;
      ***REMOVED***
      if (node.type === 'MemberExpression') {
        return node.property.name;
      // Special case for class properties
      // (babel-eslint@5 does not expose property name so we have to rely on tokens)
      ***REMOVED***
      if (node.type === 'ClassProperty') {
        const tokens = getFirstTokens(context, node, 2);
        return tokens[1] && tokens[1].type === 'Identifier' ? tokens[1].value : tokens[0].value;
      ***REMOVED***
      return '';
    ***REMOVED***

    /**
     * Checks if the Identifier node passed in looks like a defaultProps declaration.
     * @param   {ASTNode***REMOVED***  node The node to check. Must be an Identifier node.
     * @returns {boolean***REMOVED***       `true` if the node is a defaultProps declaration, `false` if not
     */
    function isDefaultPropsDeclaration(node) {
      const propName = getPropertyName(node);
      return (propName === 'defaultProps' || propName === 'getDefaultProps');
    ***REMOVED***

    function getKey(node) {
      return getText(context, node.key || node.argument);
    ***REMOVED***

    /**
     * Find a variable by name in the current scope.
     * @param  {ASTNode***REMOVED*** node The node to look for.
     * @param  {string***REMOVED*** name Name of the variable to look for.
     * @returns {ASTNode|null***REMOVED*** Return null if the variable could not be found, ASTNode otherwise.
     */
    function findVariableByName(node, name) {
      const variable = variableUtil
        .getVariableFromContext(context, node, name);

      if (!variable || !variable.defs[0] || !variable.defs[0].node) {
        return null;
      ***REMOVED***

      if (variable.defs[0].node.type === 'TypeAlias') {
        return variable.defs[0].node.right;
      ***REMOVED***

      return variable.defs[0].node.init;
    ***REMOVED***

    /**
     * Checks if defaultProps declarations are sorted
     * @param {Array***REMOVED*** declarations The array of AST nodes being checked.
     * @returns {void***REMOVED***
     */
    function checkSorted(declarations) {
      // function fix(fixer) {
      //   return propTypesSortUtil.fixPropTypesSort(context, fixer, declarations, ignoreCase);
      // ***REMOVED***

      declarations.reduce((prev, curr, idx, decls) => {
        if (/Spread(?:Property|Element)$/.test(curr.type)) {
          return decls[idx + 1];
        ***REMOVED***

        let prevPropName = getKey(prev);
        let currentPropName = getKey(curr);

        if (ignoreCase) {
          prevPropName = prevPropName.toLowerCase();
          currentPropName = currentPropName.toLowerCase();
        ***REMOVED***

        if (currentPropName < prevPropName) {
          report(context, messages.propsNotSorted, 'propsNotSorted', {
            node: curr,
            // fix
          ***REMOVED***);

          return prev;
        ***REMOVED***

        return curr;
      ***REMOVED***, declarations[0]);
    ***REMOVED***

    function checkNode(node) {
      if (!node) {
        return;
      ***REMOVED***
      if (node.type === 'ObjectExpression') {
        checkSorted(node.properties);
      ***REMOVED*** else if (node.type === 'Identifier') {
        const propTypesObject = findVariableByName(node, node.name);
        if (propTypesObject && propTypesObject.properties) {
          checkSorted(propTypesObject.properties);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    // --------------------------------------------------------------------------
    // Public API
    // --------------------------------------------------------------------------

    return {
      'ClassProperty, PropertyDefinition'(node) {
        if (!isDefaultPropsDeclaration(node)) {
          return;
        ***REMOVED***

        checkNode(node.value);
      ***REMOVED***,

      MemberExpression(node) {
        if (!isDefaultPropsDeclaration(node)) {
          return;
        ***REMOVED***

        checkNode('right' in node.parent && node.parent.right);
      ***REMOVED***,

      Program() {
        if (isWarnedForDeprecation) {
          return;
        ***REMOVED***

        log('The react/jsx-sort-default-props rule is deprecated. It has been renamed to `react/sort-default-props`.');
        isWarnedForDeprecation = true;
      ***REMOVED***,
    ***REMOVED***;
  ***REMOVED***,
***REMOVED***;
