// eslint-plugin-react/lib/rules/jsx-curly-brace-presence.js
/**
 * @fileoverview Enforce curly braces or disallow unnecessary curly brace in JSX
 * @author Jacky Ho
 * @author Simon Lydell
 */

'use strict';

const arrayIncludes = require('array-includes');

const docsUrl = require('../util/docsUrl');
const jsxUtil = require('../util/jsx');
const report = require('../util/report');
const eslintUtil = require('../util/eslint');

const getSourceCode = eslintUtil.getSourceCode;
const getText = eslintUtil.getText;

// ------------------------------------------------------------------------------
// Constants
// ------------------------------------------------------------------------------

const OPTION_ALWAYS = 'always';
const OPTION_NEVER = 'never';
const OPTION_IGNORE = 'ignore';

const OPTION_VALUES = [
  OPTION_ALWAYS,
  OPTION_NEVER,
  OPTION_IGNORE,
];
const DEFAULT_CONFIG = { props: OPTION_NEVER, children: OPTION_NEVER, propElementValues: OPTION_IGNORE ***REMOVED***;

const HTML_ENTITY_REGEX = () => /&[A-Za-z\d#]+;/g;

function containsLineTerminators(rawStringValue) {
  return /[\n\r\u2028\u2029]/.test(rawStringValue);
***REMOVED***

function containsBackslash(rawStringValue) {
  return arrayIncludes(rawStringValue, '\\');
***REMOVED***

function containsHTMLEntity(rawStringValue) {
  return HTML_ENTITY_REGEX().test(rawStringValue);
***REMOVED***

function containsOnlyHtmlEntities(rawStringValue) {
  return rawStringValue.replace(HTML_ENTITY_REGEX(), '').trim() === '';
***REMOVED***

function containsDisallowedJSXTextChars(rawStringValue) {
  return /[{<>***REMOVED***]/.test(rawStringValue);
***REMOVED***

function containsQuoteCharacters(value) {
  return /['"]/.test(value);
***REMOVED***

function containsMultilineComment(value) {
  return /\/\*/.test(value);
***REMOVED***

function escapeDoubleQuotes(rawStringValue) {
  return rawStringValue.replace(/\\"/g, '"').replace(/"/g, '\\"');
***REMOVED***

function escapeBackslashes(rawStringValue) {
  return rawStringValue.replace(/\\/g, '\\\\');
***REMOVED***

function needToEscapeCharacterForJSX(raw, node) {
  return (
    containsBackslash(raw)
    || containsHTMLEntity(raw)
    || (node.parent.type !== 'JSXAttribute' && containsDisallowedJSXTextChars(raw))
  );
***REMOVED***

function containsWhitespaceExpression(child) {
  if (child.type === 'JSXExpressionContainer') {
    const value = child.expression.value;
    return value ? jsxUtil.isWhiteSpaces(value) : false;
  ***REMOVED***
  return false;
***REMOVED***

function isLineBreak(text) {
  return containsLineTerminators(text) && text.trim() === '';
***REMOVED***

function wrapNonHTMLEntities(text) {
  const HTML_ENTITY = '<HTML_ENTITY>';
  const withCurlyBraces = text.split(HTML_ENTITY_REGEX()).map((word) => (
    word === '' ? '' : `{${JSON.stringify(word)***REMOVED******REMOVED***`
  )).join(HTML_ENTITY);

  const htmlEntities = text.match(HTML_ENTITY_REGEX());
  return htmlEntities.reduce((acc, htmlEntity) => (
    acc.replace(HTML_ENTITY, htmlEntity)
  ), withCurlyBraces);
***REMOVED***

function wrapWithCurlyBraces(rawText) {
  if (!containsLineTerminators(rawText)) {
    return `{${JSON.stringify(rawText)***REMOVED******REMOVED***`;
  ***REMOVED***

  return rawText.split('\n').map((line) => {
    if (line.trim() === '') {
      return line;
    ***REMOVED***
    const firstCharIndex = line.search(/[^\s]/);
    const leftWhitespace = line.slice(0, firstCharIndex);
    const text = line.slice(firstCharIndex);

    if (containsHTMLEntity(line)) {
      return `${leftWhitespace***REMOVED***${wrapNonHTMLEntities(text)***REMOVED***`;
    ***REMOVED***
    return `${leftWhitespace***REMOVED***{${JSON.stringify(text)***REMOVED******REMOVED***`;
  ***REMOVED***).join('\n');
***REMOVED***

function isWhiteSpaceLiteral(node) {
  return node.type && node.type === 'Literal' && node.value && jsxUtil.isWhiteSpaces(node.value);
***REMOVED***

function isStringWithTrailingWhiteSpaces(value) {
  return /^\s|\s$/.test(value);
***REMOVED***

function isLiteralWithTrailingWhiteSpaces(node) {
  return node.type && node.type === 'Literal' && node.value && isStringWithTrailingWhiteSpaces(node.value);
***REMOVED***

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

const messages = {
  unnecessaryCurly: 'Curly braces are unnecessary here.',
  missingCurly: 'Need to wrap this literal in a JSX expression.',
***REMOVED***;

/** @type {import('eslint').Rule.RuleModule***REMOVED*** */
module.exports = {
  meta: {
    docs: {
      description: 'Disallow unnecessary JSX expressions when literals alone are sufficient or enforce JSX expressions on literals in JSX children or attributes',
      category: 'Stylistic Issues',
      recommended: false,
      url: docsUrl('jsx-curly-brace-presence'),
    ***REMOVED***,
    fixable: 'code',

    messages,

    schema: [
      {
        anyOf: [
          {
            type: 'object',
            properties: {
              props: { enum: OPTION_VALUES ***REMOVED***,
              children: { enum: OPTION_VALUES ***REMOVED***,
              propElementValues: { enum: OPTION_VALUES ***REMOVED***,
            ***REMOVED***,
            additionalProperties: false,
          ***REMOVED***,
          {
            enum: OPTION_VALUES,
          ***REMOVED***,
        ],
      ***REMOVED***,
    ],
  ***REMOVED***,

  create(context) {
    const ruleOptions = context.options[0];
    const userConfig = typeof ruleOptions === 'string'
      ? { props: ruleOptions, children: ruleOptions, propElementValues: OPTION_IGNORE ***REMOVED***
      : Object.assign({***REMOVED***, DEFAULT_CONFIG, ruleOptions);

    /**
     * Report and fix an unnecessary curly brace violation on a node
     * @param {ASTNode***REMOVED*** JSXExpressionNode - The AST node with an unnecessary JSX expression
     */
    function reportUnnecessaryCurly(JSXExpressionNode) {
      report(context, messages.unnecessaryCurly, 'unnecessaryCurly', {
        node: JSXExpressionNode,
        fix(fixer) {
          const expression = JSXExpressionNode.expression;

          let textToReplace;
          if (jsxUtil.isJSX(expression)) {
            textToReplace = getText(context, expression);
          ***REMOVED*** else {
            const expressionType = expression && expression.type;
            const parentType = JSXExpressionNode.parent.type;

            if (parentType === 'JSXAttribute') {
              if (expressionType !== 'TemplateLiteral' && /["]/.test(expression.raw.slice(1, -1))) {
                textToReplace = expression.raw;
              ***REMOVED*** else {
                textToReplace = `"${expressionType === 'TemplateLiteral'
                  ? expression.quasis[0].value.raw
                  : expression.raw.slice(1, -1)
                ***REMOVED***"`;
              ***REMOVED***
            ***REMOVED*** else if (jsxUtil.isJSX(expression)) {
              textToReplace = getText(context, expression);
            ***REMOVED*** else {
              textToReplace = expressionType === 'TemplateLiteral'
                ? expression.quasis[0].value.cooked : expression.value;
            ***REMOVED***
          ***REMOVED***

          return fixer.replaceText(JSXExpressionNode, textToReplace);
        ***REMOVED***,
      ***REMOVED***);
    ***REMOVED***

    function reportMissingCurly(literalNode) {
      report(context, messages.missingCurly, 'missingCurly', {
        node: literalNode,
        fix(fixer) {
          if (jsxUtil.isJSX(literalNode)) {
            return fixer.replaceText(literalNode, `{${getText(context, literalNode)***REMOVED******REMOVED***`);
          ***REMOVED***

          // If a HTML entity name is found, bail out because it can be fixed
          // by either using the real character or the unicode equivalent.
          // If it contains any line terminator character, bail out as well.
          if (
            containsOnlyHtmlEntities(literalNode.raw)
            || (literalNode.parent.type === 'JSXAttribute' && containsLineTerminators(literalNode.raw))
            || isLineBreak(literalNode.raw)
          ) {
            return null;
          ***REMOVED***

          const expression = literalNode.parent.type === 'JSXAttribute'
            ? `{"${escapeDoubleQuotes(escapeBackslashes(
              literalNode.raw.slice(1, -1)
            ))***REMOVED***"***REMOVED***`
            : wrapWithCurlyBraces(literalNode.raw);

          return fixer.replaceText(literalNode, expression);
        ***REMOVED***,
      ***REMOVED***);
    ***REMOVED***

    // Bail out if there is any character that needs to be escaped in JSX
    // because escaping decreases readability and the original code may be more
    // readable anyway or intentional for other specific reasons
    function lintUnnecessaryCurly(JSXExpressionNode) {
      const expression = JSXExpressionNode.expression;
      const expressionType = expression.type;

      const sourceCode = getSourceCode(context);
      // Curly braces containing comments are necessary
      if (sourceCode.getCommentsInside && sourceCode.getCommentsInside(JSXExpressionNode).length > 0) {
        return;
      ***REMOVED***

      if (
        (expressionType === 'Literal' || expressionType === 'JSXText')
          && typeof expression.value === 'string'
          && (
            (JSXExpressionNode.parent.type === 'JSXAttribute' && !isWhiteSpaceLiteral(expression))
            || !isLiteralWithTrailingWhiteSpaces(expression)
          )
          && !containsMultilineComment(expression.value)
          && !needToEscapeCharacterForJSX(expression.raw, JSXExpressionNode) && (
          jsxUtil.isJSX(JSXExpressionNode.parent)
          || (!containsQuoteCharacters(expression.value) || typeof expression.value === 'string')
        )
      ) {
        reportUnnecessaryCurly(JSXExpressionNode);
      ***REMOVED*** else if (
        expressionType === 'TemplateLiteral'
        && expression.expressions.length === 0
        && expression.quasis[0].value.raw.indexOf('\n') === -1
        && !isStringWithTrailingWhiteSpaces(expression.quasis[0].value.raw)
        && !needToEscapeCharacterForJSX(expression.quasis[0].value.raw, JSXExpressionNode)
        && !containsQuoteCharacters(expression.quasis[0].value.cooked)
      ) {
        reportUnnecessaryCurly(JSXExpressionNode);
      ***REMOVED*** else if (jsxUtil.isJSX(expression)) {
        reportUnnecessaryCurly(JSXExpressionNode);
      ***REMOVED***
    ***REMOVED***

    function areRuleConditionsSatisfied(parent, config, ruleCondition) {
      return (
        parent.type === 'JSXAttribute'
          && typeof config.props === 'string'
          && config.props === ruleCondition
      ) || (
        jsxUtil.isJSX(parent)
          && typeof config.children === 'string'
          && config.children === ruleCondition
      );
    ***REMOVED***

    function getAdjacentSiblings(node, children) {
      for (let i = 1; i < children.length - 1; i++) {
        const child = children[i];
        if (node === child) {
          return [children[i - 1], children[i + 1]];
        ***REMOVED***
      ***REMOVED***
      if (node === children[0] && children[1]) {
        return [children[1]];
      ***REMOVED***
      if (node === children[children.length - 1] && children[children.length - 2]) {
        return [children[children.length - 2]];
      ***REMOVED***
      return [];
    ***REMOVED***

    function hasAdjacentJsxExpressionContainers(node, children) {
      if (!children) {
        return false;
      ***REMOVED***
      const childrenExcludingWhitespaceLiteral = children.filter((child) => !isWhiteSpaceLiteral(child));
      const adjSiblings = getAdjacentSiblings(node, childrenExcludingWhitespaceLiteral);

      return adjSiblings.some((x) => x.type && x.type === 'JSXExpressionContainer');
    ***REMOVED***
    function hasAdjacentJsx(node, children) {
      if (!children) {
        return false;
      ***REMOVED***
      const childrenExcludingWhitespaceLiteral = children.filter((child) => !isWhiteSpaceLiteral(child));
      const adjSiblings = getAdjacentSiblings(node, childrenExcludingWhitespaceLiteral);

      return adjSiblings.some((x) => x.type && arrayIncludes(['JSXExpressionContainer', 'JSXElement'], x.type));
    ***REMOVED***
    function shouldCheckForUnnecessaryCurly(node, config) {
      const parent = node.parent;
      // Bail out if the parent is a JSXAttribute & its contents aren't
      // StringLiteral or TemplateLiteral since e.g
      // <App prop1={<CustomEl />***REMOVED*** prop2={<CustomEl>...</CustomEl>***REMOVED*** />

      if (
        parent.type && parent.type === 'JSXAttribute'
        && (node.expression && node.expression.type
          && node.expression.type !== 'Literal'
          && node.expression.type !== 'StringLiteral'
          && node.expression.type !== 'TemplateLiteral')
      ) {
        return false;
      ***REMOVED***

      // If there are adjacent `JsxExpressionContainer` then there is no need,
      // to check for unnecessary curly braces.
      if (jsxUtil.isJSX(parent) && hasAdjacentJsxExpressionContainers(node, parent.children)) {
        return false;
      ***REMOVED***
      if (containsWhitespaceExpression(node) && hasAdjacentJsx(node, parent.children)) {
        return false;
      ***REMOVED***
      if (
        parent.children
        && parent.children.length === 1
        && containsWhitespaceExpression(node)
      ) {
        return false;
      ***REMOVED***

      return areRuleConditionsSatisfied(parent, config, OPTION_NEVER);
    ***REMOVED***

    function shouldCheckForMissingCurly(node, config) {
      if (jsxUtil.isJSX(node)) {
        return config.propElementValues !== OPTION_IGNORE;
      ***REMOVED***
      if (
        isLineBreak(node.raw)
        || containsOnlyHtmlEntities(node.raw)
      ) {
        return false;
      ***REMOVED***
      const parent = node.parent;
      if (
        parent.children
        && parent.children.length === 1
        && containsWhitespaceExpression(parent.children[0])
      ) {
        return false;
      ***REMOVED***

      return areRuleConditionsSatisfied(parent, config, OPTION_ALWAYS);
    ***REMOVED***

    // --------------------------------------------------------------------------
    // Public
    // --------------------------------------------------------------------------

    return {
      'JSXAttribute > JSXExpressionContainer > JSXElement'(node) {
        if (userConfig.propElementValues === OPTION_NEVER) {
          reportUnnecessaryCurly(node.parent);
        ***REMOVED***
      ***REMOVED***,

      JSXExpressionContainer(node) {
        if (shouldCheckForUnnecessaryCurly(node, userConfig)) {
          lintUnnecessaryCurly(node);
        ***REMOVED***
      ***REMOVED***,

      'JSXAttribute > JSXElement, Literal, JSXText'(node) {
        if (shouldCheckForMissingCurly(node, userConfig)) {
          reportMissingCurly(node);
        ***REMOVED***
      ***REMOVED***,
    ***REMOVED***;
  ***REMOVED***,
***REMOVED***;
