// eslint-plugin-react/lib/rules/no-find-dom-node.js
/**
 * @fileoverview Prevent usage of findDOMNode
 * @author Yannick Croissant
 */

'use strict';

const docsUrl = require('../util/docsUrl');
const report = require('../util/report');

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

const messages = {
  noFindDOMNode: 'Do not use findDOMNode. It doesn’t work with function components and is deprecated in StrictMode. See https://reactjs.org/docs/react-dom.html#finddomnode',
***REMOVED***;

/** @type {import('eslint').Rule.RuleModule***REMOVED*** */
module.exports = {
  meta: {
    docs: {
      description: 'Disallow usage of findDOMNode',
      category: 'Best Practices',
      recommended: true,
      url: docsUrl('no-find-dom-node'),
    ***REMOVED***,

    messages,

    schema: [],
  ***REMOVED***,

  create(context) {
    return {
      CallExpression(node) {
        const callee = node.callee;

        const isFindDOMNode = ('name' in callee && callee.name === 'findDOMNode') || (
          'property' in callee
          && callee.property
          && 'name' in callee.property
          && callee.property.name === 'findDOMNode'
        );

        if (!isFindDOMNode) {
          return;
        ***REMOVED***

        report(context, messages.noFindDOMNode, 'noFindDOMNode', {
          node: callee,
        ***REMOVED***);
      ***REMOVED***,
    ***REMOVED***;
  ***REMOVED***,
***REMOVED***;
