// eslint-plugin-react/lib/rules/no-unused-prop-types.js
/**
 * @fileoverview Prevent definitions of unused prop types
 * @author Evgueni Naverniouk
 */

'use strict';

const values = require('object.values');

// As for exceptions for props.children or props.className (and alike) look at
// https://github.com/jsx-eslint/eslint-plugin-react/issues/7

const Components = require('../util/Components');
const docsUrl = require('../util/docsUrl');
const report = require('../util/report');

/**
 * Checks if the component must be validated
 * @param {Object***REMOVED*** component The component to process
 * @returns {boolean***REMOVED*** True if the component must be validated, false if not.
 */
function mustBeValidated(component) {
  return !!component && !component.ignoreUnusedPropTypesValidation;
***REMOVED***

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

const messages = {
  unusedPropType: '\'{{name***REMOVED******REMOVED***\' PropType is defined but prop is never used',
***REMOVED***;

/** @type {import('eslint').Rule.RuleModule***REMOVED*** */
module.exports = {
  meta: {
    docs: {
      description: 'Disallow definitions of unused propTypes',
      category: 'Best Practices',
      recommended: false,
      url: docsUrl('no-unused-prop-types'),
    ***REMOVED***,

    messages,

    schema: [{
      type: 'object',
      properties: {
        ignore: {
          type: 'array',
          items: {
            type: 'string',
          ***REMOVED***,
          uniqueItems: true,
        ***REMOVED***,
        customValidators: {
          type: 'array',
          items: {
            type: 'string',
          ***REMOVED***,
        ***REMOVED***,
        skipShapeProps: {
          type: 'boolean',
        ***REMOVED***,
      ***REMOVED***,
      additionalProperties: false,
    ***REMOVED***],
  ***REMOVED***,

  create: Components.detect((context, components) => {
    const defaults = { skipShapeProps: true, customValidators: [], ignore: [] ***REMOVED***;
    const configuration = Object.assign({***REMOVED***, defaults, context.options[0] || {***REMOVED***);

    /**
     * Checks if the prop is ignored
     * @param {string***REMOVED*** name Name of the prop to check.
     * @returns {boolean***REMOVED*** True if the prop is ignored, false if not.
     */
    function isIgnored(name) {
      return configuration.ignore.indexOf(name) !== -1;
    ***REMOVED***

    /**
     * Checks if a prop is used
     * @param {ASTNode***REMOVED*** node The AST node being checked.
     * @param {Object***REMOVED*** prop Declared prop object
     * @returns {boolean***REMOVED*** True if the prop is used, false if not.
     */
    function isPropUsed(node, prop) {
      const usedPropTypes = node.usedPropTypes || [];
      for (let i = 0, l = usedPropTypes.length; i < l; i++) {
        const usedProp = usedPropTypes[i];
        if (
          prop.type === 'shape'
          || prop.type === 'exact'
          || prop.name === '__ANY_KEY__'
          || usedProp.name === prop.name
        ) {
          return true;
        ***REMOVED***
      ***REMOVED***

      return false;
    ***REMOVED***

    /**
     * Used to recursively loop through each declared prop type
     * @param {Object***REMOVED*** component The component to process
     * @param {ASTNode[]|true***REMOVED*** props List of props to validate
     */
    function reportUnusedPropType(component, props) {
      // Skip props that check instances
      if (props === true) {
        return;
      ***REMOVED***

      Object.keys(props || {***REMOVED***).forEach((key) => {
        const prop = props[key];
        // Skip props that check instances
        if (prop === true) {
          return;
        ***REMOVED***

        if ((prop.type === 'shape' || prop.type === 'exact') && configuration.skipShapeProps) {
          return;
        ***REMOVED***

        if (prop.node && prop.node.typeAnnotation && prop.node.typeAnnotation.typeAnnotation
          && prop.node.typeAnnotation.typeAnnotation.type === 'TSNeverKeyword') {
          return;
        ***REMOVED***

        if (prop.node && !isIgnored(prop.fullName) && !isPropUsed(component, prop)) {
          report(context, messages.unusedPropType, 'unusedPropType', {
            node: prop.node.key || prop.node,
            data: {
              name: prop.fullName,
            ***REMOVED***,
          ***REMOVED***);
        ***REMOVED***

        if (prop.children) {
          reportUnusedPropType(component, prop.children);
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***

    /**
     * Reports unused proptypes for a given component
     * @param {Object***REMOVED*** component The component to process
     */
    function reportUnusedPropTypes(component) {
      reportUnusedPropType(component, component.declaredPropTypes);
    ***REMOVED***

    // --------------------------------------------------------------------------
    // Public
    // --------------------------------------------------------------------------

    return {
      'Program:exit'() {
        // Report undeclared proptypes for all classes
        values(components.list())
          .filter((component) => mustBeValidated(component))
          .forEach((component) => {
            reportUnusedPropTypes(component);
          ***REMOVED***);
      ***REMOVED***,
    ***REMOVED***;
  ***REMOVED***),
***REMOVED***;
