// eslint-plugin-react/lib/rules/no-unstable-nested-components.js
/**
 * @fileoverview Prevent creating unstable components inside components
 * @author Ari Perkkiö
 */

'use strict';

const minimatch = require('minimatch');
const Components = require('../util/Components');
const docsUrl = require('../util/docsUrl');
const astUtil = require('../util/ast');
const isCreateElement = require('../util/isCreateElement');
const report = require('../util/report');

// ------------------------------------------------------------------------------
// Constants
// ------------------------------------------------------------------------------

const COMPONENT_AS_PROPS_INFO = ' If you want to allow component creation in props, set allowAsProps option to true.';
const HOOK_REGEXP = /^use[A-Z0-9].*$/;

// ------------------------------------------------------------------------------
// Helpers
// ------------------------------------------------------------------------------

/**
 * Generate error message with given parent component name
 * @param {string***REMOVED*** parentName Name of the parent component, if known
 * @returns {string***REMOVED*** Error message with parent component name
 */
function generateErrorMessageWithParentName(parentName) {
  return `Do not define components during render. React will see a new component type on every render and destroy the entire subtree’s DOM nodes and state (https://reactjs.org/docs/reconciliation.html#elements-of-different-types). Instead, move this component definition out of the parent component${parentName ? ` “${parentName***REMOVED***” ` : ' '***REMOVED***and pass data as props.`;
***REMOVED***

/**
 * Check whether given text matches the pattern passed in.
 * @param {string***REMOVED*** text Text to validate
 * @param {string***REMOVED*** pattern Pattern to match against
 * @returns {boolean***REMOVED***
 */
function propMatchesRenderPropPattern(text, pattern) {
  return typeof text === 'string' && minimatch(text, pattern);
***REMOVED***

/**
 * Get closest parent matching given matcher
 * @param {ASTNode***REMOVED*** node The AST node
 * @param {Context***REMOVED*** context eslint context
 * @param {Function***REMOVED*** matcher Method used to match the parent
 * @returns {ASTNode***REMOVED*** The matching parent node, if any
 */
function getClosestMatchingParent(node, context, matcher) {
  if (!node || !node.parent || node.parent.type === 'Program') {
    return;
  ***REMOVED***

  if (matcher(node.parent, context)) {
    return node.parent;
  ***REMOVED***

  return getClosestMatchingParent(node.parent, context, matcher);
***REMOVED***

/**
 * Matcher used to check whether given node is a `createElement` call
 * @param {ASTNode***REMOVED*** node The AST node
 * @param {Context***REMOVED*** context eslint context
 * @returns {boolean***REMOVED*** True if node is a `createElement` call, false if not
 */
function isCreateElementMatcher(node, context) {
  return (
    astUtil.isCallExpression(node)
    && isCreateElement(context, node)
  );
***REMOVED***

/**
 * Matcher used to check whether given node is a `ObjectExpression`
 * @param {ASTNode***REMOVED*** node The AST node
 * @returns {boolean***REMOVED*** True if node is a `ObjectExpression`, false if not
 */
function isObjectExpressionMatcher(node) {
  return node && node.type === 'ObjectExpression';
***REMOVED***

/**
 * Matcher used to check whether given node is a `JSXExpressionContainer`
 * @param {ASTNode***REMOVED*** node The AST node
 * @returns {boolean***REMOVED*** True if node is a `JSXExpressionContainer`, false if not
 */
function isJSXExpressionContainerMatcher(node) {
  return node && node.type === 'JSXExpressionContainer';
***REMOVED***

/**
 * Matcher used to check whether given node is a `JSXAttribute` of `JSXExpressionContainer`
 * @param {ASTNode***REMOVED*** node The AST node
 * @returns {boolean***REMOVED*** True if node is a `JSXAttribute` of `JSXExpressionContainer`, false if not
 */
function isJSXAttributeOfExpressionContainerMatcher(node) {
  return (
    node
    && node.type === 'JSXAttribute'
    && node.value
    && node.value.type === 'JSXExpressionContainer'
  );
***REMOVED***

/**
 * Matcher used to check whether given node is an object `Property`
 * @param {ASTNode***REMOVED*** node The AST node
 * @returns {boolean***REMOVED*** True if node is a `Property`, false if not
 */
function isPropertyOfObjectExpressionMatcher(node) {
  return (
    node
    && node.parent
    && node.parent.type === 'Property'
  );
***REMOVED***

/**
 * Check whether given node or its parent is directly inside `map` call
 * ```jsx
 * {items.map(item => <li />)***REMOVED***
 * ```
 * @param {ASTNode***REMOVED*** node The AST node
 * @returns {boolean***REMOVED*** True if node is directly inside `map` call, false if not
 */
function isMapCall(node) {
  return (
    node
    && node.callee
    && node.callee.property
    && node.callee.property.name === 'map'
  );
***REMOVED***

/**
 * Check whether given node is `ReturnStatement` of a React hook
 * @param {ASTNode***REMOVED*** node The AST node
 * @param {Context***REMOVED*** context eslint context
 * @returns {boolean***REMOVED*** True if node is a `ReturnStatement` of a React hook, false if not
 */
function isReturnStatementOfHook(node, context) {
  if (
    !node
    || !node.parent
    || node.parent.type !== 'ReturnStatement'
  ) {
    return false;
  ***REMOVED***

  const callExpression = getClosestMatchingParent(node, context, astUtil.isCallExpression);
  return (
    callExpression
    && callExpression.callee
    && HOOK_REGEXP.test(callExpression.callee.name)
  );
***REMOVED***

/**
 * Check whether given node is declared inside a render prop
 * ```jsx
 * <Component renderFooter={() => <div />***REMOVED*** />
 * <Component>{() => <div />***REMOVED***</Component>
 * ```
 * @param {ASTNode***REMOVED*** node The AST node
 * @param {Context***REMOVED*** context eslint context
 * @param {string***REMOVED*** propNamePattern a pattern to match render props against
 * @returns {boolean***REMOVED*** True if component is declared inside a render prop, false if not
 */
function isComponentInRenderProp(node, context, propNamePattern) {
  if (
    node
    && node.parent
    && node.parent.type === 'Property'
    && node.parent.key
    && propMatchesRenderPropPattern(node.parent.key.name, propNamePattern)
  ) {
    return true;
  ***REMOVED***

  // Check whether component is a render prop used as direct children, e.g. <Component>{() => <div />***REMOVED***</Component>
  if (
    node
    && node.parent
    && node.parent.type === 'JSXExpressionContainer'
    && node.parent.parent
    && node.parent.parent.type === 'JSXElement'
  ) {
    return true;
  ***REMOVED***

  const jsxExpressionContainer = getClosestMatchingParent(node, context, isJSXExpressionContainerMatcher);

  // Check whether prop name indicates accepted patterns
  if (
    jsxExpressionContainer
    && jsxExpressionContainer.parent
    && jsxExpressionContainer.parent.type === 'JSXAttribute'
    && jsxExpressionContainer.parent.name
    && jsxExpressionContainer.parent.name.type === 'JSXIdentifier'
  ) {
    const propName = jsxExpressionContainer.parent.name.name;

    // Starts with render, e.g. <Component renderFooter={() => <div />***REMOVED*** />
    if (propMatchesRenderPropPattern(propName, propNamePattern)) {
      return true;
    ***REMOVED***

    // Uses children prop explicitly, e.g. <Component children={() => <div />***REMOVED*** />
    if (propName === 'children') {
      return true;
    ***REMOVED***
  ***REMOVED***

  return false;
***REMOVED***

/**
 * Check whether given node is declared directly inside a render property
 * ```jsx
 * const rows = { render: () => <div /> ***REMOVED***
 * <Component rows={ [{ render: () => <div /> ***REMOVED***] ***REMOVED*** />
 *  ```
 * @param {ASTNode***REMOVED*** node The AST node
 * @param {string***REMOVED*** propNamePattern The pattern to match render props against
 * @returns {boolean***REMOVED*** True if component is declared inside a render property, false if not
 */
function isDirectValueOfRenderProperty(node, propNamePattern) {
  return (
    node
    && node.parent
    && node.parent.type === 'Property'
    && node.parent.key
    && node.parent.key.type === 'Identifier'
    && propMatchesRenderPropPattern(node.parent.key.name, propNamePattern)
  );
***REMOVED***

/**
 * Resolve the component name of given node
 * @param {ASTNode***REMOVED*** node The AST node of the component
 * @returns {string***REMOVED*** Name of the component, if any
 */
function resolveComponentName(node) {
  const parentName = node.id && node.id.name;
  if (parentName) return parentName;

  return (
    node.type === 'ArrowFunctionExpression'
    && node.parent
    && node.parent.id
    && node.parent.id.name
  );
***REMOVED***

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

/** @type {import('eslint').Rule.RuleModule***REMOVED*** */
module.exports = {
  meta: {
    docs: {
      description: 'Disallow creating unstable components inside components',
      category: 'Possible Errors',
      recommended: false,
      url: docsUrl('no-unstable-nested-components'),
    ***REMOVED***,
    schema: [{
      type: 'object',
      properties: {
        customValidators: {
          type: 'array',
          items: {
            type: 'string',
          ***REMOVED***,
        ***REMOVED***,
        allowAsProps: {
          type: 'boolean',
        ***REMOVED***,
        propNamePattern: {
          type: 'string',
        ***REMOVED***,
      ***REMOVED***,
      additionalProperties: false,
    ***REMOVED***],
  ***REMOVED***,

  create: Components.detect((context, components, utils) => {
    const allowAsProps = context.options.some((option) => option && option.allowAsProps);
    const propNamePattern = (context.options[0] || {***REMOVED***).propNamePattern || 'render*';

    /**
     * Check whether given node is declared inside class component's render block
     * ```jsx
     * class Component extends React.Component {
     *   render() {
     *     class NestedClassComponent extends React.Component {
     * ...
     * ```
     * @param {ASTNode***REMOVED*** node The AST node being checked
     * @returns {boolean***REMOVED*** True if node is inside class component's render block, false if not
     */
    function isInsideRenderMethod(node) {
      const parentComponent = utils.getParentComponent(node);

      if (!parentComponent || parentComponent.type !== 'ClassDeclaration') {
        return false;
      ***REMOVED***

      return (
        node
        && node.parent
        && node.parent.type === 'MethodDefinition'
        && node.parent.key
        && node.parent.key.name === 'render'
      );
    ***REMOVED***

    /**
     * Check whether given node is a function component declared inside class component.
     * Util's component detection fails to detect function components inside class components.
     * ```jsx
     * class Component extends React.Component {
     *  render() {
     *    const NestedComponent = () => <div />;
     * ...
     * ```
     * @param {ASTNode***REMOVED*** node The AST node being checked
     * @returns {boolean***REMOVED*** True if given node a function component declared inside class component, false if not
     */
    function isFunctionComponentInsideClassComponent(node) {
      const parentComponent = utils.getParentComponent(node);
      const parentStatelessComponent = utils.getParentStatelessComponent(node);

      return (
        parentComponent
        && parentStatelessComponent
        && parentComponent.type === 'ClassDeclaration'
        && utils.getStatelessComponent(parentStatelessComponent)
        && utils.isReturningJSX(node)
      );
    ***REMOVED***

    /**
     * Check whether given node is declared inside `createElement` call's props
     * ```js
     * React.createElement(Component, {
     *   footer: () => React.createElement("div", null)
     * ***REMOVED***)
     * ```
     * @param {ASTNode***REMOVED*** node The AST node
     * @returns {boolean***REMOVED*** True if node is declare inside `createElement` call's props, false if not
     */
    function isComponentInsideCreateElementsProp(node) {
      if (!components.get(node)) {
        return false;
      ***REMOVED***

      const createElementParent = getClosestMatchingParent(node, context, isCreateElementMatcher);

      return (
        createElementParent
        && createElementParent.arguments
        && createElementParent.arguments[1] === getClosestMatchingParent(node, context, isObjectExpressionMatcher)
      );
    ***REMOVED***

    /**
     * Check whether given node is declared inside a component/object prop.
     * ```jsx
     * <Component footer={() => <div />***REMOVED*** />
     * { footer: () => <div /> ***REMOVED***
     * ```
     * @param {ASTNode***REMOVED*** node The AST node being checked
     * @returns {boolean***REMOVED*** True if node is a component declared inside prop, false if not
     */
    function isComponentInProp(node) {
      if (isPropertyOfObjectExpressionMatcher(node)) {
        return utils.isReturningJSX(node);
      ***REMOVED***

      const jsxAttribute = getClosestMatchingParent(node, context, isJSXAttributeOfExpressionContainerMatcher);

      if (!jsxAttribute) {
        return isComponentInsideCreateElementsProp(node);
      ***REMOVED***

      return utils.isReturningJSX(node);
    ***REMOVED***

    /**
     * Check whether given node is a stateless component returning non-JSX
     * ```jsx
     * {{ a: () => null ***REMOVED******REMOVED***
     * ```
     * @param {ASTNode***REMOVED*** node The AST node being checked
     * @returns {boolean***REMOVED*** True if node is a stateless component returning non-JSX, false if not
     */
    function isStatelessComponentReturningNull(node) {
      const component = utils.getStatelessComponent(node);

      return component && !utils.isReturningJSX(component);
    ***REMOVED***

    /**
     * Check whether given node is a unstable nested component
     * @param {ASTNode***REMOVED*** node The AST node being checked
     */
    function validate(node) {
      if (!node || !node.parent) {
        return;
      ***REMOVED***

      const isDeclaredInsideProps = isComponentInProp(node);

      if (
        !components.get(node)
        && !isFunctionComponentInsideClassComponent(node)
        && !isDeclaredInsideProps) {
        return;
      ***REMOVED***

      if (
        // Support allowAsProps option
        (isDeclaredInsideProps && (allowAsProps || isComponentInRenderProp(node, context, propNamePattern)))

        // Prevent reporting components created inside Array.map calls
        || isMapCall(node)
        || isMapCall(node.parent)

        // Do not mark components declared inside hooks (or falsy '() => null' clean-up methods)
        || isReturnStatementOfHook(node, context)

        // Do not mark objects containing render methods
        || isDirectValueOfRenderProperty(node, propNamePattern)

        // Prevent reporting nested class components twice
        || isInsideRenderMethod(node)

        // Prevent falsely reporting detected "components" which do not return JSX
        || isStatelessComponentReturningNull(node)
      ) {
        return;
      ***REMOVED***

      // Get the closest parent component
      const parentComponent = getClosestMatchingParent(
        node,
        context,
        (nodeToMatch) => components.get(nodeToMatch)
      );

      if (parentComponent) {
        const parentName = resolveComponentName(parentComponent);

        // Exclude lowercase parents, e.g. function createTestComponent()
        // React-dom prevents creating lowercase components
        if (parentName && parentName[0] === parentName[0].toLowerCase()) {
          return;
        ***REMOVED***

        let message = generateErrorMessageWithParentName(parentName);

        // Add information about allowAsProps option when component is declared inside prop
        if (isDeclaredInsideProps && !allowAsProps) {
          message += COMPONENT_AS_PROPS_INFO;
        ***REMOVED***

        report(context, message, null, {
          node,
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***

    // --------------------------------------------------------------------------
    // Public
    // --------------------------------------------------------------------------

    return {
      FunctionDeclaration(node) { validate(node); ***REMOVED***,
      ArrowFunctionExpression(node) { validate(node); ***REMOVED***,
      FunctionExpression(node) { validate(node); ***REMOVED***,
      ClassDeclaration(node) { validate(node); ***REMOVED***,
      CallExpression(node) { validate(node); ***REMOVED***,
    ***REMOVED***;
  ***REMOVED***),
***REMOVED***;
