// eslint-plugin-react/lib/rules/no-unsafe.js
/**
 * @fileoverview Prevent usage of unsafe lifecycle methods
 * @author Sergei Startsev
 */

'use strict';

const astUtil = require('../util/ast');
const componentUtil = require('../util/componentUtil');
const docsUrl = require('../util/docsUrl');
const testReactVersion = require('../util/version').testReactVersion;
const report = require('../util/report');

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

const messages = {
  unsafeMethod: '{{method***REMOVED******REMOVED*** is unsafe for use in async rendering. Update the component to use {{newMethod***REMOVED******REMOVED*** instead. {{details***REMOVED******REMOVED***',
***REMOVED***;

/** @type {import('eslint').Rule.RuleModule***REMOVED*** */
module.exports = {
  meta: {
    docs: {
      description: 'Disallow usage of unsafe lifecycle methods',
      category: 'Best Practices',
      recommended: false,
      url: docsUrl('no-unsafe'),
    ***REMOVED***,

    messages,

    schema: [
      {
        type: 'object',
        properties: {
          checkAliases: {
            default: false,
            type: 'boolean',
          ***REMOVED***,
        ***REMOVED***,
        additionalProperties: false,
      ***REMOVED***,
    ],
  ***REMOVED***,

  create(context) {
    const config = context.options[0] || {***REMOVED***;
    const checkAliases = config.checkAliases || false;

    const isApplicable = testReactVersion(context, '>= 16.3.0');
    if (!isApplicable) {
      return {***REMOVED***;
    ***REMOVED***

    const unsafe = {
      UNSAFE_componentWillMount: {
        newMethod: 'componentDidMount',
        details: 'See https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html.',
      ***REMOVED***,
      UNSAFE_componentWillReceiveProps: {
        newMethod: 'getDerivedStateFromProps',
        details: 'See https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html.',
      ***REMOVED***,
      UNSAFE_componentWillUpdate: {
        newMethod: 'componentDidUpdate',
        details: 'See https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html.',
      ***REMOVED***,
    ***REMOVED***;
    if (checkAliases) {
      unsafe.componentWillMount = unsafe.UNSAFE_componentWillMount;
      unsafe.componentWillReceiveProps = unsafe.UNSAFE_componentWillReceiveProps;
      unsafe.componentWillUpdate = unsafe.UNSAFE_componentWillUpdate;
    ***REMOVED***

    /**
     * Returns a list of unsafe methods
     * @returns {Array***REMOVED*** A list of unsafe methods
     */
    function getUnsafeMethods() {
      return Object.keys(unsafe);
    ***REMOVED***

    /**
     * Checks if a passed method is unsafe
     * @param {string***REMOVED*** method Life cycle method
     * @returns {boolean***REMOVED*** Returns true for unsafe methods, otherwise returns false
     */
    function isUnsafe(method) {
      const unsafeMethods = getUnsafeMethods();
      return unsafeMethods.indexOf(method) !== -1;
    ***REMOVED***

    /**
     * Reports the error for an unsafe method
     * @param {ASTNode***REMOVED*** node The AST node being checked
     * @param {string***REMOVED*** method Life cycle method
     */
    function checkUnsafe(node, method) {
      if (!isUnsafe(method)) {
        return;
      ***REMOVED***

      const meta = unsafe[method];
      const newMethod = meta.newMethod;
      const details = meta.details;

      const propertyNode = astUtil.getComponentProperties(node)
        .find((property) => astUtil.getPropertyName(property) === method);

      report(context, messages.unsafeMethod, 'unsafeMethod', {
        node: propertyNode,
        data: {
          method,
          newMethod,
          details,
        ***REMOVED***,
      ***REMOVED***);
    ***REMOVED***

    /**
     * Returns life cycle methods if available
     * @param {ASTNode***REMOVED*** node The AST node being checked.
     * @returns {Array***REMOVED*** The array of methods.
     */
    function getLifeCycleMethods(node) {
      const properties = astUtil.getComponentProperties(node);
      return properties.map((property) => astUtil.getPropertyName(property));
    ***REMOVED***

    /**
     * Checks life cycle methods
     * @param {ASTNode***REMOVED*** node The AST node being checked.
     */
    function checkLifeCycleMethods(node) {
      if (componentUtil.isES5Component(node, context) || componentUtil.isES6Component(node, context)) {
        const methods = getLifeCycleMethods(node);
        methods
          .sort((a, b) => a.localeCompare(b))
          .forEach((method) => checkUnsafe(node, method));
      ***REMOVED***
    ***REMOVED***

    return {
      ClassDeclaration: checkLifeCycleMethods,
      ClassExpression: checkLifeCycleMethods,
      ObjectExpression: checkLifeCycleMethods,
    ***REMOVED***;
  ***REMOVED***,
***REMOVED***;
