// eslint-plugin-react/lib/rules/react-in-jsx-scope.js
/**
 * @fileoverview Prevent missing React when using JSX
 * @author Glen Mailer
 */

'use strict';

const variableUtil = require('../util/variable');
const pragmaUtil = require('../util/pragma');
const docsUrl = require('../util/docsUrl');
const report = require('../util/report');

// -----------------------------------------------------------------------------
// Rule Definition
// -----------------------------------------------------------------------------

const messages = {
  notInScope: '\'{{name***REMOVED******REMOVED***\' must be in scope when using JSX',
***REMOVED***;

/** @type {import('eslint').Rule.RuleModule***REMOVED*** */
module.exports = {
  meta: {
    docs: {
      description: 'Disallow missing React when using JSX',
      category: 'Possible Errors',
      recommended: true,
      url: docsUrl('react-in-jsx-scope'),
    ***REMOVED***,

    messages,

    schema: [],
  ***REMOVED***,

  create(context) {
    const pragma = pragmaUtil.getFromContext(context);

    function checkIfReactIsInScope(node) {
      if (variableUtil.getVariableFromContext(context, node, pragma)) {
        return;
      ***REMOVED***
      report(context, messages.notInScope, 'notInScope', {
        node,
        data: {
          name: pragma,
        ***REMOVED***,
      ***REMOVED***);
    ***REMOVED***

    return {
      JSXOpeningElement: checkIfReactIsInScope,
      JSXOpeningFragment: checkIfReactIsInScope,
    ***REMOVED***;
  ***REMOVED***,
***REMOVED***;
