// eslint-plugin-react/lib/rules/sort-prop-types.js
/**
 * @fileoverview Enforce propTypes declarations alphabetical sorting
 */

'use strict';

const astUtil = require('../util/ast');
const variableUtil = require('../util/variable');
const propsUtil = require('../util/props');
const docsUrl = require('../util/docsUrl');
const propWrapperUtil = require('../util/propWrapper');
const propTypesSortUtil = require('../util/propTypesSort');
const report = require('../util/report');
const eslintUtil = require('../util/eslint');

const getSourceCode = eslintUtil.getSourceCode;
const getText = eslintUtil.getText;

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

const messages = {
  requiredPropsFirst: 'Required prop types must be listed before all other prop types',
  callbackPropsLast: 'Callback prop types must be listed after all other prop types',
  propsNotSorted: 'Prop types declarations should be sorted alphabetically',
***REMOVED***;

function getKey(context, node) {
  if (node.type === 'ObjectTypeProperty') {
    return getSourceCode(context).getFirstToken(node).value;
  ***REMOVED***
  if (node.key && node.key.value) {
    return node.key.value;
  ***REMOVED***
  return getText(context, node.key || node.argument);
***REMOVED***

/** @type {import('eslint').Rule.RuleModule***REMOVED*** */
module.exports = {
  meta: {
    docs: {
      description: 'Enforce propTypes declarations alphabetical sorting',
      category: 'Stylistic Issues',
      recommended: false,
      url: docsUrl('sort-prop-types'),
    ***REMOVED***,
    fixable: 'code',

    messages,

    schema: [{
      type: 'object',
      properties: {
        requiredFirst: {
          type: 'boolean',
        ***REMOVED***,
        callbacksLast: {
          type: 'boolean',
        ***REMOVED***,
        ignoreCase: {
          type: 'boolean',
        ***REMOVED***,
        // Whether alphabetical sorting should be enforced
        noSortAlphabetically: {
          type: 'boolean',
        ***REMOVED***,
        sortShapeProp: {
          type: 'boolean',
        ***REMOVED***,
        checkTypes: {
          type: 'boolean',
        ***REMOVED***,
      ***REMOVED***,
      additionalProperties: false,
    ***REMOVED***],
  ***REMOVED***,

  create(context) {
    const configuration = context.options[0] || {***REMOVED***;
    const requiredFirst = configuration.requiredFirst || false;
    const callbacksLast = configuration.callbacksLast || false;
    const ignoreCase = configuration.ignoreCase || false;
    const noSortAlphabetically = configuration.noSortAlphabetically || false;
    const sortShapeProp = configuration.sortShapeProp || false;
    const checkTypes = configuration.checkTypes || false;

    const typeAnnotations = new Map();

    /**
     * Checks if propTypes declarations are sorted
     * @param {Array***REMOVED*** declarations The array of AST nodes being checked.
     * @returns {void***REMOVED***
     */
    function checkSorted(declarations) {
      // Declarations will be `undefined` if the `shape` is not a literal. For
      // example, if it is a propType imported from another file.
      if (!declarations) {
        return;
      ***REMOVED***

      function fix(fixer) {
        return propTypesSortUtil.fixPropTypesSort(
          context,
          fixer,
          declarations,
          ignoreCase,
          requiredFirst,
          callbacksLast,
          noSortAlphabetically,
          sortShapeProp,
          checkTypes
        );
      ***REMOVED***

      const callbackPropsLastSeen = new WeakSet();
      const requiredPropsFirstSeen = new WeakSet();
      const propsNotSortedSeen = new WeakSet();

      declarations.reduce((prev, curr, idx, decls) => {
        if (curr.type === 'ExperimentalSpreadProperty' || curr.type === 'SpreadElement') {
          return decls[idx + 1];
        ***REMOVED***

        let prevPropName = getKey(context, prev);
        let currentPropName = getKey(context, curr);
        const previousIsRequired = propTypesSortUtil.isRequiredProp(prev);
        const currentIsRequired = propTypesSortUtil.isRequiredProp(curr);
        const previousIsCallback = propTypesSortUtil.isCallbackPropName(prevPropName);
        const currentIsCallback = propTypesSortUtil.isCallbackPropName(currentPropName);

        if (ignoreCase) {
          prevPropName = String(prevPropName).toLowerCase();
          currentPropName = String(currentPropName).toLowerCase();
        ***REMOVED***

        if (requiredFirst) {
          if (previousIsRequired && !currentIsRequired) {
            // Transition between required and non-required. Don't compare for alphabetical.
            return curr;
          ***REMOVED***
          if (!previousIsRequired && currentIsRequired) {
            // Encountered a non-required prop after a required prop
            if (!requiredPropsFirstSeen.has(curr)) {
              requiredPropsFirstSeen.add(curr);
              report(context, messages.requiredPropsFirst, 'requiredPropsFirst', {
                node: curr,
                fix,
              ***REMOVED***);
            ***REMOVED***
            return curr;
          ***REMOVED***
        ***REMOVED***

        if (callbacksLast) {
          if (!previousIsCallback && currentIsCallback) {
            // Entering the callback prop section
            return curr;
          ***REMOVED***
          if (previousIsCallback && !currentIsCallback) {
            // Encountered a non-callback prop after a callback prop
            if (!callbackPropsLastSeen.has(prev)) {
              callbackPropsLastSeen.add(prev);
              report(context, messages.callbackPropsLast, 'callbackPropsLast', {
                node: prev,
                fix,
              ***REMOVED***);
            ***REMOVED***
            return prev;
          ***REMOVED***
        ***REMOVED***

        if (!noSortAlphabetically && currentPropName < prevPropName) {
          if (!propsNotSortedSeen.has(curr)) {
            propsNotSortedSeen.add(curr);
            report(context, messages.propsNotSorted, 'propsNotSorted', {
              node: curr,
              fix,
            ***REMOVED***);
          ***REMOVED***
          return prev;
        ***REMOVED***

        return curr;
      ***REMOVED***, declarations[0]);
    ***REMOVED***

    function checkNode(node) {
      if (!node) {
        return;
      ***REMOVED***

      if (node.type === 'ObjectExpression') {
        checkSorted(node.properties);
      ***REMOVED*** else if (node.type === 'Identifier') {
        const propTypesObject = variableUtil.findVariableByName(context, node, node.name);
        if (propTypesObject && propTypesObject.properties) {
          checkSorted(propTypesObject.properties);
        ***REMOVED***
      ***REMOVED*** else if (astUtil.isCallExpression(node)) {
        const innerNode = node.arguments && node.arguments[0];
        if (propWrapperUtil.isPropWrapperFunction(context, node.callee.name) && innerNode) {
          checkNode(innerNode);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    function handleFunctionComponent(node) {
      const firstArg = node.params
        && node.params.length > 0
        && node.params[0].typeAnnotation
        && node.params[0].typeAnnotation.typeAnnotation;
      if (firstArg && firstArg.type === 'TSTypeReference') {
        const propType = typeAnnotations.get(firstArg.typeName.name)
          && typeAnnotations.get(firstArg.typeName.name)[0];
        if (propType && propType.members) {
          checkSorted(propType.members);
        ***REMOVED***
      ***REMOVED*** else if (firstArg && firstArg.type === 'TSTypeLiteral') {
        if (firstArg.members) {
          checkSorted(firstArg.members);
        ***REMOVED***
      ***REMOVED*** else if (firstArg && firstArg.type === 'GenericTypeAnnotation') {
        const propType = typeAnnotations.get(firstArg.id.name)
          && typeAnnotations.get(firstArg.id.name)[0];
        if (propType && propType.properties) {
          checkSorted(propType.properties);
        ***REMOVED***
      ***REMOVED*** else if (firstArg && firstArg.type === 'ObjectTypeAnnotation') {
        if (firstArg.properties) {
          checkSorted(firstArg.properties);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    return Object.assign({
      CallExpression(node) {
        if (!sortShapeProp || !propTypesSortUtil.isShapeProp(node) || !(node.arguments && node.arguments[0])) {
          return;
        ***REMOVED***

        const firstArg = node.arguments[0];
        if (firstArg.properties) {
          checkSorted(firstArg.properties);
        ***REMOVED*** else if (firstArg.type === 'Identifier') {
          const variable = variableUtil.findVariableByName(context, node, firstArg.name);
          if (variable && variable.properties) {
            checkSorted(variable.properties);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***,

      'ClassProperty, PropertyDefinition'(node) {
        if (!propsUtil.isPropTypesDeclaration(node)) {
          return;
        ***REMOVED***
        checkNode(node.value);
      ***REMOVED***,

      MemberExpression(node) {
        if (!propsUtil.isPropTypesDeclaration(node)) {
          return;
        ***REMOVED***

        checkNode(node.parent.right);
      ***REMOVED***,

      ObjectExpression(node) {
        node.properties.forEach((property) => {
          if (!property.key) {
            return;
          ***REMOVED***

          if (!propsUtil.isPropTypesDeclaration(property)) {
            return;
          ***REMOVED***
          if (property.value.type === 'ObjectExpression') {
            checkSorted(property.value.properties);
          ***REMOVED***
        ***REMOVED***);
      ***REMOVED***,
    ***REMOVED***, checkTypes ? {
      TSTypeLiteral(node) {
        if (node && node.parent.id) {
          const currentNode = [].concat(
            typeAnnotations.get(node.parent.id.name) || [],
            node
          );
          typeAnnotations.set(node.parent.id.name, currentNode);
        ***REMOVED***
      ***REMOVED***,

      TypeAlias(node) {
        if (node.right.type === 'ObjectTypeAnnotation') {
          const currentNode = [].concat(
            typeAnnotations.get(node.id.name) || [],
            node.right
          );
          typeAnnotations.set(node.id.name, currentNode);
        ***REMOVED***
      ***REMOVED***,

      TSTypeAliasDeclaration(node) {
        if (node.typeAnnotation.type === 'TSTypeLiteral' || node.typeAnnotation.type === 'ObjectTypeAnnotation') {
          const currentNode = [].concat(
            typeAnnotations.get(node.id.name) || [],
            node.typeAnnotation
          );
          typeAnnotations.set(node.id.name, currentNode);
        ***REMOVED***
      ***REMOVED***,
      FunctionDeclaration: handleFunctionComponent,
      ArrowFunctionExpression: handleFunctionComponent,
    ***REMOVED*** : null);
  ***REMOVED***,
***REMOVED***;
