---
description: 'Require explicit return types on functions and class methods.'
---

> üõë This file is source code, not the primary documentation location! üõë
>
> See **https://typescript-eslint.io/rules/explicit-function-return-type** for documentation.

Functions in TypeScript often don't need to be given an explicit return type annotation.
Leaving off the return type is less code to read or write and allows the compiler to infer it from the contents of the function.

However, explicit return types do make it visually more clear what type is returned by a function.
They can also speed up TypeScript type checking performance in large codebases with many large functions.

This rule enforces that functions do have an explicit return type annotation.

## Examples

<!--tabs-->

### ‚ùå Incorrect

```ts
// Should indicate that no value is returned (void)
function test() {
  return;
***REMOVED***

// Should indicate that a number is returned
var fn = function () {
  return 1;
***REMOVED***;

// Should indicate that a string is returned
var arrowFn = () => 'test';

class Test {
  // Should indicate that no value is returned (void)
  method() {
    return;
  ***REMOVED***
***REMOVED***
```

### ‚úÖ Correct

```ts
// No return value should be expected (void)
function test(): void {
  return;
***REMOVED***

// A return value of type number
var fn = function (): number {
  return 1;
***REMOVED***;

// A return value of type string
var arrowFn = (): string => 'test';

class Test {
  // No return value should be expected (void)
  method(): void {
    return;
  ***REMOVED***
***REMOVED***
```

## Options

### Configuring in a mixed JS/TS codebase

If you are working on a codebase within which you lint non-TypeScript code (i.e. `.js`/`.mjs`/`.cjs`/`.jsx`), you should ensure that you should use [ESLint `overrides`](https://eslint.org/docs/user-guide/configuring#disabling-rules-only-for-a-group-of-files) to only enable the rule on `.ts`/`.mts`/`.cts`/`.tsx` files. If you don't, then you will get unfixable lint errors reported within `.js`/`.mjs`/`.cjs`/`.jsx` files.

```jsonc
{
  "rules": {
    // disable the rule for all files
    "@typescript-eslint/explicit-function-return-type": "off"
  ***REMOVED***,
  "overrides": [
    {
      // enable the rule specifically for TypeScript files
      "files": ["*.ts", "*.mts", "*.cts", "*.tsx"],
      "rules": {
        "@typescript-eslint/explicit-function-return-type": "error"
      ***REMOVED***
    ***REMOVED***
  ]
***REMOVED***
```

### `allowExpressions`

Examples of code for this rule with `{ allowExpressions: true ***REMOVED***`:

<!--tabs-->

#### ‚ùå Incorrect

```ts
function test() {***REMOVED***

const fn = () => {***REMOVED***;

export default () => {***REMOVED***;
```

#### ‚úÖ Correct

```ts
node.addEventListener('click', () => {***REMOVED***);

node.addEventListener('click', function () {***REMOVED***);

const foo = arr.map(i => i * i);
```

### `allowTypedFunctionExpressions`

Examples of code for this rule with `{ allowTypedFunctionExpressions: true ***REMOVED***`:

<!--tabs-->

#### ‚ùå Incorrect

```ts
let arrowFn = () => 'test';

let funcExpr = function () {
  return 'test';
***REMOVED***;

let objectProp = {
  foo: () => 1,
***REMOVED***;
```

#### ‚úÖ Correct

```ts
type FuncType = () => string;

let arrowFn: FuncType = () => 'test';

let funcExpr: FuncType = function() {
  return 'test';
***REMOVED***;

let asTyped = (() => '') as () => string;
let castTyped = <() => string>(() => '');

interface ObjectType {
  foo(): number;
***REMOVED***
let objectProp: ObjectType = {
  foo: () => 1,
***REMOVED***;
let objectPropAs = {
  foo: () => 1,
***REMOVED*** as ObjectType;
let objectPropCast = <ObjectType>{
  foo: () => 1,
***REMOVED***;

declare functionWithArg(arg: () => number);
functionWithArg(() => 1);

declare functionWithObjectArg(arg: { method: () => number ***REMOVED***);
functionWithObjectArg({
  method() {
    return 1;
  ***REMOVED***,
***REMOVED***);
```

### `allowHigherOrderFunctions`

Examples of code for this rule with `{ allowHigherOrderFunctions: true ***REMOVED***`:

<!--tabs-->

#### ‚ùå Incorrect

```ts
var arrowFn = () => () => {***REMOVED***;

function fn() {
  return function () {***REMOVED***;
***REMOVED***
```

#### ‚úÖ Correct

```ts
var arrowFn = () => (): void => {***REMOVED***;

function fn() {
  return function (): void {***REMOVED***;
***REMOVED***
```

### `allowDirectConstAssertionInArrowFunctions`

Examples of code for this rule with `{ allowDirectConstAssertionInArrowFunctions: true ***REMOVED***`:

<!--tabs-->

#### ‚ùå Incorrect

```ts
const func = (value: number) => ({ type: 'X', value ***REMOVED*** as any);
const func = (value: number) => ({ type: 'X', value ***REMOVED*** as Action);
```

#### ‚úÖ Correct

```ts
const func = (value: number) => ({ foo: 'bar', value ***REMOVED*** as const);
const func = () => x as const;
```

### `allowConciseArrowFunctionExpressionsStartingWithVoid`

Examples of code for this rule with `{ allowConciseArrowFunctionExpressionsStartingWithVoid: true ***REMOVED***`:

<!--tabs-->

#### ‚ùå Incorrect

```ts
var join = (a: string, b: string) => `${a***REMOVED***${b***REMOVED***`;

const log = (message: string) => {
  console.log(message);
***REMOVED***;
```

#### ‚úÖ Correct

```ts
var log = (message: string) => void console.log(message);
```

### `allowFunctionsWithoutTypeParameters`

Examples of code for this rule with `{ allowFunctionsWithoutTypeParameters: true ***REMOVED***`:

<!--tabs-->

#### ‚ùå Incorrect

```ts
function foo<T>(t: T) {
  return t;
***REMOVED***

const bar = <T>(t: T) => t;
```

#### ‚úÖ Correct

```ts
function foo<T>(t: T): T {
  return t;
***REMOVED***

const bar = <T>(t: T): T => t;

const allowedFunction(x: string) {
  return x;
***REMOVED***

const allowedArrow = (x: string) => x;
```

### `allowedNames`

You may pass function/method names you would like this rule to ignore, like so:

```json
{
  "@typescript-eslint/explicit-function-return-type": [
    "error",
    {
      "allowedNames": ["ignoredFunctionName", "ignoredMethodName"]
    ***REMOVED***
  ]
***REMOVED***
```

### `allowIIFE`

Examples of code for this rule with `{ allowIIFE: true ***REMOVED***`:

#### ‚ùå Incorrect

```ts
var func = () => 'foo';
```

#### ‚úÖ Correct

```ts
var foo = (() => 'foo')();

var bar = (function () {
  return 'bar';
***REMOVED***)();
```

## When Not To Use It

If you don't wish to prevent calling code from using function return values in unexpected ways, then
you will not need this rule.

## Further Reading

- TypeScript [Functions](https://www.typescriptlang.org/docs/handbook/functions.html#function-types)
