---
description: 'Disallow the declaration of empty interfaces.'
---

> üõë This file is source code, not the primary documentation location! üõë
>
> See **https://typescript-eslint.io/rules/no-empty-interface** for documentation.

An empty interface in TypeScript does very little: any non-nullable value is assignable to `{***REMOVED***`.
Using an empty interface is often a sign of programmer error, such as misunderstanding the concept of `{***REMOVED***` or forgetting to fill in fields.

This rule aims to ensure that only meaningful interfaces are declared in the code.

## Examples

<!--tabs-->

### ‚ùå Incorrect

```ts
// an empty interface
interface Foo {***REMOVED***

// an interface with only one supertype (Bar === Foo)
interface Bar extends Foo {***REMOVED***

// an interface with an empty list of supertypes
interface Baz {***REMOVED***
```

### ‚úÖ Correct

```ts
// an interface with any number of members
interface Foo {
  name: string;
***REMOVED***

// same as above
interface Bar {
  age: number;
***REMOVED***

// an interface with more than one supertype
// in this case the interface can be used as a replacement of an intersection type.
interface Baz extends Foo, Bar {***REMOVED***
```

<!--/tabs-->

## Options

This rule accepts a single object option with the following default configuration:

```json
{
  "@typescript-eslint/no-empty-interface": [
    "error",
    {
      "allowSingleExtends": false
    ***REMOVED***
  ]
***REMOVED***
```

- `allowSingleExtends: true` will silence warnings about extending a single interface without adding additional members

## When Not To Use It

If you don't care about having empty/meaningless interfaces, then you will not need this rule.
