---
description: 'Disallow TypeScript namespaces.'
---

> üõë This file is source code, not the primary documentation location! üõë
>
> See **https://typescript-eslint.io/rules/no-namespace** for documentation.

TypeScript historically allowed a form of code organization called "custom modules" (`module Example {***REMOVED***`), later renamed to "namespaces" (`namespace Example`).
Namespaces are an outdated way to organize TypeScript code.
ES2015 module syntax is now preferred (`import`/`export`).

> This rule does not report on the use of TypeScript module declarations to describe external APIs (`declare module 'foo' {***REMOVED***`).

## Examples

Examples of code with the default options:

<!--tabs-->

### ‚ùå Incorrect

```ts
module foo {***REMOVED***
namespace foo {***REMOVED***

declare module foo {***REMOVED***
declare namespace foo {***REMOVED***
```

### ‚úÖ Correct

```ts
declare module 'foo' {***REMOVED***

// anything inside a d.ts file
```

<!--/tabs-->

## Options

### `allowDeclarations`

Examples of code with the `{ "allowDeclarations": true ***REMOVED***` option:

<!--tabs-->

#### ‚ùå Incorrect

```ts
module foo {***REMOVED***
namespace foo {***REMOVED***
```

#### ‚úÖ Correct

```ts
declare module 'foo' {***REMOVED***
declare module foo {***REMOVED***
declare namespace foo {***REMOVED***

declare global {
  namespace foo {***REMOVED***
***REMOVED***

declare module foo {
  namespace foo {***REMOVED***
***REMOVED***
```

<!--/tabs-->

Examples of code for the `{ "allowDeclarations": false ***REMOVED***` option:

<!--tabs-->

#### ‚ùå Incorrect

```ts
module foo {***REMOVED***
namespace foo {***REMOVED***
declare module foo {***REMOVED***
declare namespace foo {***REMOVED***
```

#### ‚úÖ Correct

```ts
declare module 'foo' {***REMOVED***
```

### `allowDefinitionFiles`

Examples of code for the `{ "allowDefinitionFiles": true ***REMOVED***` option:

<!--tabs-->

#### ‚ùå Incorrect

```ts
// if outside a d.ts file
module foo {***REMOVED***
namespace foo {***REMOVED***

// if outside a d.ts file and allowDeclarations = false
module foo {***REMOVED***
namespace foo {***REMOVED***
declare module foo {***REMOVED***
declare namespace foo {***REMOVED***
```

#### ‚úÖ Correct

```ts
declare module 'foo' {***REMOVED***

// anything inside a d.ts file
```

## When Not To Use It

If you are using the ES2015 module syntax, then you will not need this rule.

## Further Reading

- [Modules](https://www.typescriptlang.org/docs/handbook/modules.html)
- [Namespaces](https://www.typescriptlang.org/docs/handbook/namespaces.html)
- [Namespaces and Modules](https://www.typescriptlang.org/docs/handbook/namespaces-and-modules.html)
