---
description: 'Disallow specified modules when loaded by `import`.'
---

> üõë This file is source code, not the primary documentation location! üõë
>
> See **https://typescript-eslint.io/rules/no-restricted-imports** for documentation.

## Examples

This rule extends the base [`eslint/no-restricted-imports`](https://eslint.org/docs/rules/no-restricted-imports) rule.

## Options

This rule adds the following options:

### `allowTypeImports`

(default: `false`)

You can specify this option for a specific path or pattern as follows:

```jsonc
"@typescript-eslint/no-restricted-imports": ["error", {
  "paths": [{
    "name": "import-foo",
    "message": "Please use import-bar instead.",
    "allowTypeImports": true
  ***REMOVED***, {
    "name": "import-baz",
    "message": "Please use import-quux instead.",
    "allowTypeImports": true
  ***REMOVED***]
***REMOVED***]
```

When set to `true`, the rule will allow [Type-Only Imports](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export).

Examples of code with the above config:

<!--tabs-->

#### ‚ùå Incorrect

```ts
import foo from 'import-foo';
export { Foo ***REMOVED*** from 'import-foo';

import baz from 'import-baz';
export { Baz ***REMOVED*** from 'import-baz';
```

#### ‚úÖ Correct

```ts
import { foo ***REMOVED*** from 'other-module';

import type foo from 'import-foo';
export type { Foo ***REMOVED*** from 'import-foo';

import type baz from 'import-baz';
export type { Baz ***REMOVED*** from 'import-baz';
```
