---
description: 'Disallow type arguments that are equal to the default.'
---

> üõë This file is source code, not the primary documentation location! üõë
>
> See **https://typescript-eslint.io/rules/no-unnecessary-type-arguments** for documentation.

Type parameters in TypeScript may specify a default value.
For example:

```ts
function f<T = number>(...) {...***REMOVED***
```

It is redundant to provide an explicit type parameter equal to that default: e.g. calling `f<number>(...)`.
This rule reports when an explicitly specified type argument is the default for that type parameter.

## Examples

<!--tabs-->

### ‚ùå Incorrect

```ts
function f<T = number>() {***REMOVED***
f<number>();
```

```ts
function g<T = number, U = string>() {***REMOVED***
g<string, string>();
```

```ts
class C<T = number> {***REMOVED***
new C<number>();

class D extends C<number> {***REMOVED***
```

```ts
interface I<T = number> {***REMOVED***
class Impl implements I<number> {***REMOVED***
```

### ‚úÖ Correct

```ts
function f<T = number>() {***REMOVED***
f();
f<string>();
```

```ts
function g<T = number, U = string>() {***REMOVED***
g<string>();
g<number, number>();
```

```ts
class C<T = number> {***REMOVED***
new C();
new C<string>();

class D extends C {***REMOVED***
class D extends C<string> {***REMOVED***
```

```ts
interface I<T = number> {***REMOVED***
class Impl implements I<string> {***REMOVED***
```
