---
description: 'Disallow unsafe declaration merging.'
---

> üõë This file is source code, not the primary documentation location! üõë
>
> See **https://typescript-eslint.io/rules/no-unsafe-declaration-merging** for documentation.

TypeScript's "declaration merging" supports merging separate declarations with the same name.

Declaration merging between classes and interfaces is unsafe.
The TypeScript compiler doesn't check whether properties are initialized, which can cause lead to TypeScript not detecting code that will cause runtime errors.

```ts
interface Foo {
  nums: number[];
***REMOVED***

class Foo {***REMOVED***

const foo = new Foo();

foo.nums.push(1); // Runtime Error: Cannot read properties of undefined.
```

## Examples

<!--tabs-->

### ‚ùå Incorrect

```ts
interface Foo {***REMOVED***

class Foo {***REMOVED***
```

### ‚úÖ Correct

```ts
interface Foo {***REMOVED***
class Bar implements Foo {***REMOVED***

namespace Baz {***REMOVED***
namespace Baz {***REMOVED***
enum Baz {***REMOVED***

namespace Qux {***REMOVED***
function Qux() {***REMOVED***
```

## Further Reading

- [Declaration Merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html)
