---
description: 'Require function parameters to be typed as `readonly` to prevent accidental mutation of inputs.'
---

> üõë This file is source code, not the primary documentation location! üõë
>
> See **https://typescript-eslint.io/rules/prefer-readonly-parameter-types** for documentation.

Mutating function arguments can lead to confusing, hard to debug behavior.
Whilst it's easy to implicitly remember to not modify function arguments, explicitly typing arguments as readonly provides clear contract to consumers.
This contract makes it easier for a consumer to reason about if a function has side-effects.

This rule allows you to enforce that function parameters resolve to readonly types.
A type is considered readonly if:

- it is a primitive type (`string`, `number`, `boolean`, `symbol`, or an enum),
- it is a function signature type,
- it is a readonly array type whose element type is considered readonly.
- it is a readonly tuple type whose elements are all considered readonly.
- it is an object type whose properties are all marked as readonly, and whose values are all considered readonly.

## Examples

<!--tabs-->

### ‚ùå Incorrect

```ts
function array1(arg: string[]) {***REMOVED*** // array is not readonly
function array2(arg: readonly string[][]) {***REMOVED*** // array element is not readonly
function array3(arg: [string, number]) {***REMOVED*** // tuple is not readonly
function array4(arg: readonly [string[], number]) {***REMOVED*** // tuple element is not readonly
// the above examples work the same if you use ReadonlyArray<T> instead

function object1(arg: { prop: string ***REMOVED***) {***REMOVED*** // property is not readonly
function object2(arg: { readonly prop: string; prop2: string ***REMOVED***) {***REMOVED*** // not all properties are readonly
function object3(arg: { readonly prop: { prop2: string ***REMOVED*** ***REMOVED***) {***REMOVED*** // nested property is not readonly
// the above examples work the same if you use Readonly<T> instead

interface CustomArrayType extends ReadonlyArray<string> {
  prop: string; // note: this property is mutable
***REMOVED***
function custom1(arg: CustomArrayType) {***REMOVED***

interface CustomFunction {
  (): void;
  prop: string; // note: this property is mutable
***REMOVED***
function custom2(arg: CustomFunction) {***REMOVED***

function union(arg: string[] | ReadonlyArray<number[]>) {***REMOVED*** // not all types are readonly

// rule also checks function types
interface Foo {
  (arg: string[]): void;
***REMOVED***
interface Foo {
  new (arg: string[]): void;
***REMOVED***
const x = { foo(arg: string[]): void; ***REMOVED***;
function foo(arg: string[]);
type Foo = (arg: string[]) => void;
interface Foo {
  foo(arg: string[]): void;
***REMOVED***
```

### ‚úÖ Correct

```ts
function array1(arg: readonly string[]) {***REMOVED***
function array2(arg: readonly (readonly string[])[]) {***REMOVED***
function array3(arg: readonly [string, number]) {***REMOVED***
function array4(arg: readonly [readonly string[], number]) {***REMOVED***
// the above examples work the same if you use ReadonlyArray<T> instead

function object1(arg: { readonly prop: string ***REMOVED***) {***REMOVED***
function object2(arg: { readonly prop: string; readonly prop2: string ***REMOVED***) {***REMOVED***
function object3(arg: { readonly prop: { readonly prop2: string ***REMOVED*** ***REMOVED***) {***REMOVED***
// the above examples work the same if you use Readonly<T> instead

interface CustomArrayType extends ReadonlyArray<string> {
  readonly prop: string;
***REMOVED***
function custom1(arg: Readonly<CustomArrayType>) {***REMOVED***
// interfaces that extend the array types are not considered arrays, and thus must be made readonly.

interface CustomFunction {
  (): void;
  readonly prop: string;
***REMOVED***
function custom2(arg: CustomFunction) {***REMOVED***

function union(arg: readonly string[] | ReadonlyArray<number[]>) {***REMOVED***

function primitive1(arg: string) {***REMOVED***
function primitive2(arg: number) {***REMOVED***
function primitive3(arg: boolean) {***REMOVED***
function primitive4(arg: unknown) {***REMOVED***
function primitive5(arg: null) {***REMOVED***
function primitive6(arg: undefined) {***REMOVED***
function primitive7(arg: any) {***REMOVED***
function primitive8(arg: never) {***REMOVED***
function primitive9(arg: string | number | undefined) {***REMOVED***

function fnSig(arg: () => void) {***REMOVED***

enum Foo { a, b ***REMOVED***
function enum(arg: Foo) {***REMOVED***

function symb1(arg: symbol) {***REMOVED***
const customSymbol = Symbol('a');
function symb2(arg: typeof customSymbol) {***REMOVED***

// function types
interface Foo {
  (arg: readonly string[]): void;
***REMOVED***
interface Foo {
  new (arg: readonly string[]): void;
***REMOVED***
const x = { foo(arg: readonly string[]): void; ***REMOVED***;
function foo(arg: readonly string[]);
type Foo = (arg: readonly string[]) => void;
interface Foo {
  foo(arg: readonly string[]): void;
***REMOVED***
```

## Options

### `checkParameterProperties`

This option allows you to enable or disable the checking of parameter properties.
Because parameter properties create properties on the class, it may be undesirable to force them to be readonly.

Examples of code for this rule with `{checkParameterProperties: true***REMOVED***`:

<!--tabs-->

#### ‚ùå Incorrect

```ts
class Foo {
  constructor(private paramProp: string[]) {***REMOVED***
***REMOVED***
```

#### ‚úÖ Correct

```ts
class Foo {
  constructor(private paramProp: readonly string[]) {***REMOVED***
***REMOVED***
```

<!--/tabs-->

Examples of **correct** code for this rule with `{checkParameterProperties: false***REMOVED***`:

```ts
class Foo {
  constructor(
    private paramProp1: string[],
    private paramProp2: readonly string[],
  ) {***REMOVED***
***REMOVED***
```

### `ignoreInferredTypes`

This option allows you to ignore parameters which don't explicitly specify a type. This may be desirable in cases where an external dependency specifies a callback with mutable parameters, and manually annotating the callback's parameters is undesirable.

Examples of code for this rule with `{ignoreInferredTypes: true***REMOVED***`:

<!--tabs-->

#### ‚ùå Incorrect

```ts
import { acceptsCallback, CallbackOptions ***REMOVED*** from 'external-dependency';

acceptsCallback((options: CallbackOptions) => {***REMOVED***);
```

<details>
<summary>external-dependency.d.ts</summary>

```ts
export interface CallbackOptions {
  prop: string;
***REMOVED***
type Callback = (options: CallbackOptions) => void;
type AcceptsCallback = (callback: Callback) => void;

export const acceptsCallback: AcceptsCallback;
```

</details>

#### ‚úÖ Correct

```ts
import { acceptsCallback ***REMOVED*** from 'external-dependency';

acceptsCallback(options => {***REMOVED***);
```

<details>
<summary>external-dependency.d.ts</summary>

```ts
export interface CallbackOptions {
  prop: string;
***REMOVED***
type Callback = (options: CallbackOptions) => void;
type AcceptsCallback = (callback: Callback) => void;

export const acceptsCallback: AcceptsCallback;
```

</details>

### `treatMethodsAsReadonly`

This option allows you to treat all mutable methods as though they were readonly. This may be desirable when you are never reassigning methods.

Examples of code for this rule with `{treatMethodsAsReadonly: false***REMOVED***`:

<!--tabs-->

#### ‚ùå Incorrect

```ts
type MyType = {
  readonly prop: string;
  method(): string; // note: this method is mutable
***REMOVED***;
function foo(arg: MyType) {***REMOVED***
```

#### ‚úÖ Correct

```ts
type MyType = Readonly<{
  prop: string;
  method(): string;
***REMOVED***>;
function foo(arg: MyType) {***REMOVED***

type MyOtherType = {
  readonly prop: string;
  readonly method: () => string;
***REMOVED***;
function bar(arg: MyOtherType) {***REMOVED***
```

<!--/tabs-->

Examples of **correct** code for this rule with `{treatMethodsAsReadonly: true***REMOVED***`:

```ts
type MyType = {
  readonly prop: string;
  method(): string; // note: this method is mutable
***REMOVED***;
function foo(arg: MyType) {***REMOVED***
```
