// webpack/lib/dependencies/AMDDefineDependencyParserPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const RuntimeGlobals = require("../RuntimeGlobals");
const AMDDefineDependency = require("./AMDDefineDependency");
const AMDRequireArrayDependency = require("./AMDRequireArrayDependency");
const AMDRequireContextDependency = require("./AMDRequireContextDependency");
const AMDRequireItemDependency = require("./AMDRequireItemDependency");
const ConstDependency = require("./ConstDependency");
const ContextDependencyHelpers = require("./ContextDependencyHelpers");
const DynamicExports = require("./DynamicExports");
const LocalModuleDependency = require("./LocalModuleDependency");
const { addLocalModule, getLocalModule ***REMOVED*** = require("./LocalModulesHelpers");

/** @typedef {import("estree").ArrowFunctionExpression***REMOVED*** ArrowFunctionExpression */
/** @typedef {import("estree").CallExpression***REMOVED*** CallExpression */
/** @typedef {import("estree").Expression***REMOVED*** Expression */
/** @typedef {import("estree").FunctionExpression***REMOVED*** FunctionExpression */
/** @typedef {import("estree").Identifier***REMOVED*** Identifier */
/** @typedef {import("estree").Literal***REMOVED*** Literal */
/** @typedef {import("estree").MemberExpression***REMOVED*** MemberExpression */
/** @typedef {import("estree").ObjectExpression***REMOVED*** ObjectExpression */
/** @typedef {import("estree").SimpleCallExpression***REMOVED*** SimpleCallExpression */
/** @typedef {import("estree").SpreadElement***REMOVED*** SpreadElement */
/** @typedef {import("../../declarations/WebpackOptions").JavascriptParserOptions***REMOVED*** JavascriptParserOptions */
/** @typedef {import("../Dependency").DependencyLocation***REMOVED*** DependencyLocation */
/** @typedef {import("../javascript/BasicEvaluatedExpression")***REMOVED*** BasicEvaluatedExpression */
/** @typedef {import("../javascript/JavascriptParser")***REMOVED*** JavascriptParser */
/** @typedef {import("../javascript/JavascriptParser").Range***REMOVED*** Range */

/**
 * @param {Expression | SpreadElement***REMOVED*** expr expression
 * @returns {expr is CallExpression***REMOVED*** true if it's a bound function expression
 */
const isBoundFunctionExpression = expr => {
	if (expr.type !== "CallExpression") return false;
	if (expr.callee.type !== "MemberExpression") return false;
	if (expr.callee.computed) return false;
	if (expr.callee.object.type !== "FunctionExpression") return false;
	if (expr.callee.property.type !== "Identifier") return false;
	if (expr.callee.property.name !== "bind") return false;
	return true;
***REMOVED***;

/** @typedef {FunctionExpression | ArrowFunctionExpression***REMOVED*** UnboundFunctionExpression */

/**
 * @param {Expression | SpreadElement***REMOVED*** expr expression
 * @returns {expr is FunctionExpression | ArrowFunctionExpression***REMOVED*** true when unbound function expression
 */
const isUnboundFunctionExpression = expr => {
	if (expr.type === "FunctionExpression") return true;
	if (expr.type === "ArrowFunctionExpression") return true;
	return false;
***REMOVED***;

/**
 * @param {Expression | SpreadElement***REMOVED*** expr expression
 * @returns {expr is FunctionExpression | ArrowFunctionExpression | CallExpression***REMOVED*** true when callable
 */
const isCallable = expr => {
	if (isUnboundFunctionExpression(expr)) return true;
	if (isBoundFunctionExpression(expr)) return true;
	return false;
***REMOVED***;

class AMDDefineDependencyParserPlugin {
	/**
	 * @param {JavascriptParserOptions***REMOVED*** options parserOptions
	 */
	constructor(options) {
		this.options = options;
	***REMOVED***

	/**
	 * @param {JavascriptParser***REMOVED*** parser the parser
	 * @returns {void***REMOVED***
	 */
	apply(parser) {
		parser.hooks.call
			.for("define")
			.tap(
				"AMDDefineDependencyParserPlugin",
				this.processCallDefine.bind(this, parser)
			);
	***REMOVED***

	/**
	 * @param {JavascriptParser***REMOVED*** parser the parser
	 * @param {CallExpression***REMOVED*** expr call expression
	 * @param {BasicEvaluatedExpression***REMOVED*** param param
	 * @param {Record<number, string>***REMOVED*** identifiers identifiers
	 * @param {string=***REMOVED*** namedModule named module
	 * @returns {boolean | undefined***REMOVED*** result
	 */
	processArray(parser, expr, param, identifiers, namedModule) {
		if (param.isArray()) {
			const items = /** @type {BasicEvaluatedExpression[]***REMOVED*** */ (param.items);
			for (const [idx, item] of items.entries()) {
				if (
					item.isString() &&
					["require", "module", "exports"].includes(
						/** @type {string***REMOVED*** */ (item.string)
					)
				)
					identifiers[/** @type {number***REMOVED*** */ (idx)] = /** @type {string***REMOVED*** */ (
						item.string
					);
				const result = this.processItem(parser, expr, item, namedModule);
				if (result === undefined) {
					this.processContext(parser, expr, item);
				***REMOVED***
			***REMOVED***
			return true;
		***REMOVED*** else if (param.isConstArray()) {
			/** @type {(string | LocalModuleDependency | AMDRequireItemDependency)[]***REMOVED*** */
			const deps = [];
			const array = /** @type {string[]***REMOVED*** */ (param.array);
			for (const [idx, request] of array.entries()) {
				let dep;
				let localModule;
				if (request === "require") {
					identifiers[idx] = request;
					dep = RuntimeGlobals.require;
				***REMOVED*** else if (["exports", "module"].includes(request)) {
					identifiers[idx] = request;
					dep = request;
				***REMOVED*** else if ((localModule = getLocalModule(parser.state, request))) {
					localModule.flagUsed();
					dep = new LocalModuleDependency(localModule, undefined, false);
					dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
					parser.state.module.addPresentationalDependency(dep);
				***REMOVED*** else {
					dep = this.newRequireItemDependency(request);
					dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
					dep.optional = Boolean(parser.scope.inTry);
					parser.state.current.addDependency(dep);
				***REMOVED***
				deps.push(dep);
			***REMOVED***
			const dep = this.newRequireArrayDependency(
				deps,
				/** @type {Range***REMOVED*** */ (param.range)
			);
			dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
			dep.optional = Boolean(parser.scope.inTry);
			parser.state.module.addPresentationalDependency(dep);
			return true;
		***REMOVED***
	***REMOVED***

	/**
	 * @param {JavascriptParser***REMOVED*** parser the parser
	 * @param {CallExpression***REMOVED*** expr call expression
	 * @param {BasicEvaluatedExpression***REMOVED*** param param
	 * @param {string=***REMOVED*** namedModule named module
	 * @returns {boolean | undefined***REMOVED*** result
	 */
	processItem(parser, expr, param, namedModule) {
		if (param.isConditional()) {
			const options = /** @type {BasicEvaluatedExpression[]***REMOVED*** */ (param.options);
			for (const item of options) {
				const result = this.processItem(parser, expr, item);
				if (result === undefined) {
					this.processContext(parser, expr, item);
				***REMOVED***
			***REMOVED***

			return true;
		***REMOVED*** else if (param.isString()) {
			let dep;
			let localModule;

			if (param.string === "require") {
				dep = new ConstDependency(
					RuntimeGlobals.require,
					/** @type {Range***REMOVED*** */ (param.range),
					[RuntimeGlobals.require]
				);
			***REMOVED*** else if (param.string === "exports") {
				dep = new ConstDependency(
					"exports",
					/** @type {Range***REMOVED*** */ (param.range),
					[RuntimeGlobals.exports]
				);
			***REMOVED*** else if (param.string === "module") {
				dep = new ConstDependency(
					"module",
					/** @type {Range***REMOVED*** */ (param.range),
					[RuntimeGlobals.module]
				);
			***REMOVED*** else if (
				(localModule = getLocalModule(
					parser.state,
					/** @type {string***REMOVED*** */ (param.string),
					namedModule
				))
			) {
				localModule.flagUsed();
				dep = new LocalModuleDependency(localModule, param.range, false);
			***REMOVED*** else {
				dep = this.newRequireItemDependency(
					/** @type {string***REMOVED*** */ (param.string),
					param.range
				);
				dep.optional = Boolean(parser.scope.inTry);
				parser.state.current.addDependency(dep);
				return true;
			***REMOVED***
			dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
			parser.state.module.addPresentationalDependency(dep);
			return true;
		***REMOVED***
	***REMOVED***

	/**
	 * @param {JavascriptParser***REMOVED*** parser the parser
	 * @param {CallExpression***REMOVED*** expr call expression
	 * @param {BasicEvaluatedExpression***REMOVED*** param param
	 * @returns {boolean | undefined***REMOVED*** result
	 */
	processContext(parser, expr, param) {
		const dep = ContextDependencyHelpers.create(
			AMDRequireContextDependency,
			/** @type {Range***REMOVED*** */ (param.range),
			param,
			expr,
			this.options,
			{
				category: "amd"
			***REMOVED***,
			parser
		);
		if (!dep) return;
		dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
		dep.optional = Boolean(parser.scope.inTry);
		parser.state.current.addDependency(dep);
		return true;
	***REMOVED***

	/**
	 * @param {JavascriptParser***REMOVED*** parser the parser
	 * @param {CallExpression***REMOVED*** expr call expression
	 * @returns {boolean | undefined***REMOVED*** result
	 */
	processCallDefine(parser, expr) {
		/** @type {Expression | SpreadElement | undefined***REMOVED*** */
		let array;
		/** @type {FunctionExpression | ArrowFunctionExpression | CallExpression | Identifier | undefined***REMOVED*** */
		let fn;
		/** @type {ObjectExpression | Identifier | undefined***REMOVED*** */
		let obj;
		/** @type {string | undefined***REMOVED*** */
		let namedModule;
		switch (expr.arguments.length) {
			case 1:
				if (isCallable(expr.arguments[0])) {
					// define(f() {…***REMOVED***)
					fn = expr.arguments[0];
				***REMOVED*** else if (expr.arguments[0].type === "ObjectExpression") {
					// define({…***REMOVED***)
					obj = expr.arguments[0];
				***REMOVED*** else {
					// define(expr)
					// unclear if function or object
					obj = fn = /** @type {Identifier***REMOVED*** */ (expr.arguments[0]);
				***REMOVED***
				break;
			case 2:
				if (expr.arguments[0].type === "Literal") {
					namedModule = /** @type {string***REMOVED*** */ (expr.arguments[0].value);
					// define("…", …)
					if (isCallable(expr.arguments[1])) {
						// define("…", f() {…***REMOVED***)
						fn = expr.arguments[1];
					***REMOVED*** else if (expr.arguments[1].type === "ObjectExpression") {
						// define("…", {…***REMOVED***)
						obj = expr.arguments[1];
					***REMOVED*** else {
						// define("…", expr)
						// unclear if function or object
						obj = fn = /** @type {Identifier***REMOVED*** */ (expr.arguments[1]);
					***REMOVED***
				***REMOVED*** else {
					array = expr.arguments[0];
					if (isCallable(expr.arguments[1])) {
						// define([…], f() {***REMOVED***)
						fn = expr.arguments[1];
					***REMOVED*** else if (expr.arguments[1].type === "ObjectExpression") {
						// define([…], {…***REMOVED***)
						obj = expr.arguments[1];
					***REMOVED*** else {
						// define([…], expr)
						// unclear if function or object
						obj = fn = /** @type {Identifier***REMOVED*** */ (expr.arguments[1]);
					***REMOVED***
				***REMOVED***
				break;
			case 3:
				// define("…", […], f() {…***REMOVED***)
				namedModule =
					/** @type {string***REMOVED*** */
					(
						/** @type {Literal***REMOVED*** */
						(expr.arguments[0]).value
					);
				array = expr.arguments[1];
				if (isCallable(expr.arguments[2])) {
					// define("…", […], f() {***REMOVED***)
					fn = expr.arguments[2];
				***REMOVED*** else if (expr.arguments[2].type === "ObjectExpression") {
					// define("…", […], {…***REMOVED***)
					obj = expr.arguments[2];
				***REMOVED*** else {
					// define("…", […], expr)
					// unclear if function or object
					obj = fn = /** @type {Identifier***REMOVED*** */ (expr.arguments[2]);
				***REMOVED***
				break;
			default:
				return;
		***REMOVED***
		DynamicExports.bailout(parser.state);
		/** @type {Identifier[] | null***REMOVED*** */
		let fnParams = null;
		let fnParamsOffset = 0;
		if (fn) {
			if (isUnboundFunctionExpression(fn)) {
				fnParams =
					/** @type {Identifier[]***REMOVED*** */
					(fn.params);
			***REMOVED*** else if (isBoundFunctionExpression(fn)) {
				const object =
					/** @type {FunctionExpression***REMOVED*** */
					(/** @type {MemberExpression***REMOVED*** */ (fn.callee).object);

				fnParams =
					/** @type {Identifier[]***REMOVED*** */
					(object.params);
				fnParamsOffset = fn.arguments.length - 1;
				if (fnParamsOffset < 0) {
					fnParamsOffset = 0;
				***REMOVED***
			***REMOVED***
		***REMOVED***
		const fnRenames = new Map();
		if (array) {
			/** @type {Record<number, string>***REMOVED*** */
			const identifiers = {***REMOVED***;
			const param = parser.evaluateExpression(array);
			const result = this.processArray(
				parser,
				expr,
				param,
				identifiers,
				namedModule
			);
			if (!result) return;
			if (fnParams) {
				fnParams = fnParams.slice(fnParamsOffset).filter((param, idx) => {
					if (identifiers[idx]) {
						fnRenames.set(param.name, parser.getVariableInfo(identifiers[idx]));
						return false;
					***REMOVED***
					return true;
				***REMOVED***);
			***REMOVED***
		***REMOVED*** else {
			const identifiers = ["require", "exports", "module"];
			if (fnParams) {
				fnParams = fnParams.slice(fnParamsOffset).filter((param, idx) => {
					if (identifiers[idx]) {
						fnRenames.set(param.name, parser.getVariableInfo(identifiers[idx]));
						return false;
					***REMOVED***
					return true;
				***REMOVED***);
			***REMOVED***
		***REMOVED***
		/** @type {boolean | undefined***REMOVED*** */
		let inTry;
		if (fn && isUnboundFunctionExpression(fn)) {
			inTry = parser.scope.inTry;
			parser.inScope(/** @type {Identifier[]***REMOVED*** */ (fnParams), () => {
				for (const [name, varInfo] of fnRenames) {
					parser.setVariable(name, varInfo);
				***REMOVED***
				parser.scope.inTry = /** @type {boolean***REMOVED*** */ (inTry);
				if (fn.body.type === "BlockStatement") {
					parser.detectMode(fn.body.body);
					const prev = parser.prevStatement;
					parser.preWalkStatement(fn.body);
					parser.prevStatement = prev;
					parser.walkStatement(fn.body);
				***REMOVED*** else {
					parser.walkExpression(fn.body);
				***REMOVED***
			***REMOVED***);
		***REMOVED*** else if (fn && isBoundFunctionExpression(fn)) {
			inTry = parser.scope.inTry;

			const object =
				/** @type {FunctionExpression***REMOVED*** */
				(/** @type {MemberExpression***REMOVED*** */ (fn.callee).object);

			parser.inScope(
				/** @type {Identifier[]***REMOVED*** */
				(object.params).filter(
					i => !["require", "module", "exports"].includes(i.name)
				),
				() => {
					for (const [name, varInfo] of fnRenames) {
						parser.setVariable(name, varInfo);
					***REMOVED***
					parser.scope.inTry = /** @type {boolean***REMOVED*** */ (inTry);

					if (object.body.type === "BlockStatement") {
						parser.detectMode(object.body.body);
						const prev = parser.prevStatement;
						parser.preWalkStatement(object.body);
						parser.prevStatement = prev;
						parser.walkStatement(object.body);
					***REMOVED*** else {
						parser.walkExpression(
							/** @type {TODO***REMOVED*** */
							(object.body)
						);
					***REMOVED***
				***REMOVED***
			);
			if (fn.arguments) {
				parser.walkExpressions(fn.arguments);
			***REMOVED***
		***REMOVED*** else if (fn || obj) {
			parser.walkExpression(
				/** @type {FunctionExpression | ArrowFunctionExpression | CallExpression | ObjectExpression | Identifier***REMOVED*** */
				(fn || obj)
			);
		***REMOVED***

		const dep = this.newDefineDependency(
			/** @type {Range***REMOVED*** */ (expr.range),
			array ? /** @type {Range***REMOVED*** */ (array.range) : null,
			fn ? /** @type {Range***REMOVED*** */ (fn.range) : null,
			obj ? /** @type {Range***REMOVED*** */ (obj.range) : null,
			namedModule || null
		);
		dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
		if (namedModule) {
			dep.localModule = addLocalModule(parser.state, namedModule);
		***REMOVED***
		parser.state.module.addPresentationalDependency(dep);
		return true;
	***REMOVED***

	/**
	 * @param {Range***REMOVED*** range range
	 * @param {Range | null***REMOVED*** arrayRange array range
	 * @param {Range | null***REMOVED*** functionRange function range
	 * @param {Range | null***REMOVED*** objectRange object range
	 * @param {string | null***REMOVED*** namedModule true, when define is called with a name
	 * @returns {AMDDefineDependency***REMOVED*** AMDDefineDependency
	 */
	newDefineDependency(
		range,
		arrayRange,
		functionRange,
		objectRange,
		namedModule
	) {
		return new AMDDefineDependency(
			range,
			arrayRange,
			functionRange,
			objectRange,
			namedModule
		);
	***REMOVED***

	/**
	 * @param {(string | LocalModuleDependency | AMDRequireItemDependency)[]***REMOVED*** depsArray deps array
	 * @param {Range***REMOVED*** range range
	 * @returns {AMDRequireArrayDependency***REMOVED*** AMDRequireArrayDependency
	 */
	newRequireArrayDependency(depsArray, range) {
		return new AMDRequireArrayDependency(depsArray, range);
	***REMOVED***

	/**
	 * @param {string***REMOVED*** request request
	 * @param {Range=***REMOVED*** range range
	 * @returns {AMDRequireItemDependency***REMOVED*** AMDRequireItemDependency
	 */
	newRequireItemDependency(request, range) {
		return new AMDRequireItemDependency(request, range);
	***REMOVED***
***REMOVED***

module.exports = AMDDefineDependencyParserPlugin;
