// webpack/lib/dependencies/AMDRequireDependenciesBlockParserPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const RuntimeGlobals = require("../RuntimeGlobals");
const UnsupportedFeatureWarning = require("../UnsupportedFeatureWarning");
const AMDRequireArrayDependency = require("./AMDRequireArrayDependency");
const AMDRequireContextDependency = require("./AMDRequireContextDependency");
const AMDRequireDependenciesBlock = require("./AMDRequireDependenciesBlock");
const AMDRequireDependency = require("./AMDRequireDependency");
const AMDRequireItemDependency = require("./AMDRequireItemDependency");
const ConstDependency = require("./ConstDependency");
const ContextDependencyHelpers = require("./ContextDependencyHelpers");
const LocalModuleDependency = require("./LocalModuleDependency");
const { getLocalModule ***REMOVED*** = require("./LocalModulesHelpers");
const UnsupportedDependency = require("./UnsupportedDependency");
const getFunctionExpression = require("./getFunctionExpression");

/** @typedef {import("estree").CallExpression***REMOVED*** CallExpression */
/** @typedef {import("estree").Expression***REMOVED*** Expression */
/** @typedef {import("estree").Identifier***REMOVED*** Identifier */
/** @typedef {import("estree").SourceLocation***REMOVED*** SourceLocation */
/** @typedef {import("estree").SpreadElement***REMOVED*** SpreadElement */
/** @typedef {import("../../declarations/WebpackOptions").JavascriptParserOptions***REMOVED*** JavascriptParserOptions */
/** @typedef {import("../Dependency").DependencyLocation***REMOVED*** DependencyLocation */
/** @typedef {import("../Module").BuildInfo***REMOVED*** BuildInfo */
/** @typedef {import("../javascript/BasicEvaluatedExpression")***REMOVED*** BasicEvaluatedExpression */
/** @typedef {import("../javascript/JavascriptParser")***REMOVED*** JavascriptParser */
/** @typedef {import("../javascript/JavascriptParser").Range***REMOVED*** Range */

class AMDRequireDependenciesBlockParserPlugin {
	/**
	 * @param {JavascriptParserOptions***REMOVED*** options parserOptions
	 */
	constructor(options) {
		this.options = options;
	***REMOVED***

	/**
	 * @param {JavascriptParser***REMOVED*** parser the parser
	 * @param {Expression | SpreadElement***REMOVED*** expression expression
	 * @returns {boolean***REMOVED*** need bind this
	 */
	processFunctionArgument(parser, expression) {
		let bindThis = true;
		const fnData = getFunctionExpression(expression);
		if (fnData) {
			parser.inScope(
				fnData.fn.params.filter(
					i =>
						!["require", "module", "exports"].includes(
							/** @type {Identifier***REMOVED*** */ (i).name
						)
				),
				() => {
					if (fnData.fn.body.type === "BlockStatement") {
						parser.walkStatement(fnData.fn.body);
					***REMOVED*** else {
						parser.walkExpression(fnData.fn.body);
					***REMOVED***
				***REMOVED***
			);
			parser.walkExpressions(fnData.expressions);
			if (fnData.needThis === false) {
				bindThis = false;
			***REMOVED***
		***REMOVED*** else {
			parser.walkExpression(expression);
		***REMOVED***
		return bindThis;
	***REMOVED***

	/**
	 * @param {JavascriptParser***REMOVED*** parser the parser
	 * @returns {void***REMOVED***
	 */
	apply(parser) {
		parser.hooks.call
			.for("require")
			.tap(
				"AMDRequireDependenciesBlockParserPlugin",
				this.processCallRequire.bind(this, parser)
			);
	***REMOVED***

	/**
	 * @param {JavascriptParser***REMOVED*** parser the parser
	 * @param {CallExpression***REMOVED*** expr call expression
	 * @param {BasicEvaluatedExpression***REMOVED*** param param
	 * @returns {boolean | undefined***REMOVED*** result
	 */
	processArray(parser, expr, param) {
		if (param.isArray()) {
			for (const p of /** @type {BasicEvaluatedExpression[]***REMOVED*** */ (param.items)) {
				const result = this.processItem(parser, expr, p);
				if (result === undefined) {
					this.processContext(parser, expr, p);
				***REMOVED***
			***REMOVED***
			return true;
		***REMOVED*** else if (param.isConstArray()) {
			/** @type {(string | LocalModuleDependency | AMDRequireItemDependency)[]***REMOVED*** */
			const deps = [];
			for (const request of /** @type {EXPECTED_ANY[]***REMOVED*** */ (param.array)) {
				let dep;
				let localModule;
				if (request === "require") {
					dep = RuntimeGlobals.require;
				***REMOVED*** else if (["exports", "module"].includes(request)) {
					dep = request;
				***REMOVED*** else if ((localModule = getLocalModule(parser.state, request))) {
					localModule.flagUsed();
					dep = new LocalModuleDependency(localModule, undefined, false);
					dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
					parser.state.module.addPresentationalDependency(dep);
				***REMOVED*** else {
					dep = this.newRequireItemDependency(request);
					dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
					dep.optional = Boolean(parser.scope.inTry);
					parser.state.current.addDependency(dep);
				***REMOVED***
				deps.push(dep);
			***REMOVED***
			const dep = this.newRequireArrayDependency(
				deps,
				/** @type {Range***REMOVED*** */ (param.range)
			);
			dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
			dep.optional = Boolean(parser.scope.inTry);
			parser.state.module.addPresentationalDependency(dep);
			return true;
		***REMOVED***
	***REMOVED***

	/**
	 * @param {JavascriptParser***REMOVED*** parser the parser
	 * @param {CallExpression***REMOVED*** expr call expression
	 * @param {BasicEvaluatedExpression***REMOVED*** param param
	 * @returns {boolean | undefined***REMOVED*** result
	 */
	processItem(parser, expr, param) {
		if (param.isConditional()) {
			for (const p of /** @type {BasicEvaluatedExpression[]***REMOVED*** */ (
				param.options
			)) {
				const result = this.processItem(parser, expr, p);
				if (result === undefined) {
					this.processContext(parser, expr, p);
				***REMOVED***
			***REMOVED***
			return true;
		***REMOVED*** else if (param.isString()) {
			let dep;
			let localModule;
			if (param.string === "require") {
				dep = new ConstDependency(
					RuntimeGlobals.require,
					/** @type {TODO***REMOVED*** */
					(param.string),
					[RuntimeGlobals.require]
				);
			***REMOVED*** else if (param.string === "module") {
				dep = new ConstDependency(
					/** @type {string***REMOVED*** */
					(
						/** @type {BuildInfo***REMOVED*** */
						(parser.state.module.buildInfo).moduleArgument
					),
					/** @type {Range***REMOVED*** */ (param.range),
					[RuntimeGlobals.module]
				);
			***REMOVED*** else if (param.string === "exports") {
				dep = new ConstDependency(
					/** @type {string***REMOVED*** */
					(
						/** @type {BuildInfo***REMOVED*** */
						(parser.state.module.buildInfo).exportsArgument
					),
					/** @type {Range***REMOVED*** */ (param.range),
					[RuntimeGlobals.exports]
				);
			***REMOVED*** else if (
				(localModule = getLocalModule(
					parser.state,
					/** @type {string***REMOVED*** */ (param.string)
				))
			) {
				localModule.flagUsed();
				dep = new LocalModuleDependency(localModule, param.range, false);
			***REMOVED*** else {
				dep = this.newRequireItemDependency(
					/** @type {string***REMOVED*** */ (param.string),
					param.range
				);
				dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
				dep.optional = Boolean(parser.scope.inTry);
				parser.state.current.addDependency(dep);
				return true;
			***REMOVED***
			dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
			parser.state.module.addPresentationalDependency(dep);
			return true;
		***REMOVED***
	***REMOVED***

	/**
	 * @param {JavascriptParser***REMOVED*** parser the parser
	 * @param {CallExpression***REMOVED*** expr call expression
	 * @param {BasicEvaluatedExpression***REMOVED*** param param
	 * @returns {boolean | undefined***REMOVED*** result
	 */
	processContext(parser, expr, param) {
		const dep = ContextDependencyHelpers.create(
			AMDRequireContextDependency,
			/** @type {Range***REMOVED*** */
			(param.range),
			param,
			expr,
			this.options,
			{
				category: "amd"
			***REMOVED***,
			parser
		);
		if (!dep) return;
		dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
		dep.optional = Boolean(parser.scope.inTry);
		parser.state.current.addDependency(dep);
		return true;
	***REMOVED***

	/**
	 * @param {BasicEvaluatedExpression***REMOVED*** param param
	 * @returns {string | undefined***REMOVED*** result
	 */
	processArrayForRequestString(param) {
		if (param.isArray()) {
			const result =
				/** @type {BasicEvaluatedExpression[]***REMOVED*** */
				(param.items).map(item => this.processItemForRequestString(item));
			if (result.every(Boolean)) return result.join(" ");
		***REMOVED*** else if (param.isConstArray()) {
			return /** @type {string[]***REMOVED*** */ (param.array).join(" ");
		***REMOVED***
	***REMOVED***

	/**
	 * @param {BasicEvaluatedExpression***REMOVED*** param param
	 * @returns {string | undefined***REMOVED*** result
	 */
	processItemForRequestString(param) {
		if (param.isConditional()) {
			const result =
				/** @type {BasicEvaluatedExpression[]***REMOVED*** */
				(param.options).map(item => this.processItemForRequestString(item));
			if (result.every(Boolean)) return result.join("|");
		***REMOVED*** else if (param.isString()) {
			return param.string;
		***REMOVED***
	***REMOVED***

	/**
	 * @param {JavascriptParser***REMOVED*** parser the parser
	 * @param {CallExpression***REMOVED*** expr call expression
	 * @returns {boolean | undefined***REMOVED*** result
	 */
	processCallRequire(parser, expr) {
		/** @type {BasicEvaluatedExpression | undefined***REMOVED*** */
		let param;
		/** @type {AMDRequireDependenciesBlock | undefined | null***REMOVED*** */
		let depBlock;
		/** @type {AMDRequireDependency | undefined***REMOVED*** */
		let dep;
		/** @type {boolean | undefined***REMOVED*** */
		let result;

		const old = parser.state.current;

		if (expr.arguments.length >= 1) {
			param = parser.evaluateExpression(
				/** @type {Expression***REMOVED*** */ (expr.arguments[0])
			);
			depBlock = this.newRequireDependenciesBlock(
				/** @type {DependencyLocation***REMOVED*** */ (expr.loc),
				this.processArrayForRequestString(param)
			);
			dep = this.newRequireDependency(
				/** @type {Range***REMOVED*** */ (expr.range),
				/** @type {Range***REMOVED*** */ (param.range),
				expr.arguments.length > 1
					? /** @type {Range***REMOVED*** */ (expr.arguments[1].range)
					: null,
				expr.arguments.length > 2
					? /** @type {Range***REMOVED*** */ (expr.arguments[2].range)
					: null
			);
			dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
			depBlock.addDependency(dep);

			parser.state.current = /** @type {TODO***REMOVED*** */ (depBlock);
		***REMOVED***

		if (expr.arguments.length === 1) {
			parser.inScope([], () => {
				result = this.processArray(
					parser,
					expr,
					/** @type {BasicEvaluatedExpression***REMOVED*** */ (param)
				);
			***REMOVED***);
			parser.state.current = old;
			if (!result) return;
			parser.state.current.addBlock(
				/** @type {AMDRequireDependenciesBlock***REMOVED*** */ (depBlock)
			);
			return true;
		***REMOVED***

		if (expr.arguments.length === 2 || expr.arguments.length === 3) {
			try {
				parser.inScope([], () => {
					result = this.processArray(
						parser,
						expr,
						/** @type {BasicEvaluatedExpression***REMOVED*** */ (param)
					);
				***REMOVED***);
				if (!result) {
					const dep = new UnsupportedDependency(
						"unsupported",
						/** @type {Range***REMOVED*** */ (expr.range)
					);
					old.addPresentationalDependency(dep);
					if (parser.state.module) {
						parser.state.module.addError(
							new UnsupportedFeatureWarning(
								`Cannot statically analyse 'require(…, …)' in line ${
									/** @type {SourceLocation***REMOVED*** */ (expr.loc).start.line
								***REMOVED***`,
								/** @type {DependencyLocation***REMOVED*** */ (expr.loc)
							)
						);
					***REMOVED***
					depBlock = null;
					return true;
				***REMOVED***
				/** @type {AMDRequireDependency***REMOVED*** */
				(dep).functionBindThis = this.processFunctionArgument(
					parser,
					expr.arguments[1]
				);
				if (expr.arguments.length === 3) {
					/** @type {AMDRequireDependency***REMOVED*** */
					(dep).errorCallbackBindThis = this.processFunctionArgument(
						parser,
						expr.arguments[2]
					);
				***REMOVED***
			***REMOVED*** finally {
				parser.state.current = old;
				if (depBlock) parser.state.current.addBlock(depBlock);
			***REMOVED***
			return true;
		***REMOVED***
	***REMOVED***

	/**
	 * @param {DependencyLocation***REMOVED*** loc location
	 * @param {string=***REMOVED*** request request
	 * @returns {AMDRequireDependenciesBlock***REMOVED*** AMDRequireDependenciesBlock
	 */
	newRequireDependenciesBlock(loc, request) {
		return new AMDRequireDependenciesBlock(loc, request);
	***REMOVED***

	/**
	 * @param {Range***REMOVED*** outerRange outer range
	 * @param {Range***REMOVED*** arrayRange array range
	 * @param {Range | null***REMOVED*** functionRange function range
	 * @param {Range | null***REMOVED*** errorCallbackRange error callback range
	 * @returns {AMDRequireDependency***REMOVED*** dependency
	 */
	newRequireDependency(
		outerRange,
		arrayRange,
		functionRange,
		errorCallbackRange
	) {
		return new AMDRequireDependency(
			outerRange,
			arrayRange,
			functionRange,
			errorCallbackRange
		);
	***REMOVED***

	/**
	 * @param {string***REMOVED*** request request
	 * @param {Range=***REMOVED*** range range
	 * @returns {AMDRequireItemDependency***REMOVED*** AMDRequireItemDependency
	 */
	newRequireItemDependency(request, range) {
		return new AMDRequireItemDependency(request, range);
	***REMOVED***

	/**
	 * @param {(string | LocalModuleDependency | AMDRequireItemDependency)[]***REMOVED*** depsArray deps array
	 * @param {Range***REMOVED*** range range
	 * @returns {AMDRequireArrayDependency***REMOVED*** AMDRequireArrayDependency
	 */
	newRequireArrayDependency(depsArray, range) {
		return new AMDRequireArrayDependency(depsArray, range);
	***REMOVED***
***REMOVED***
module.exports = AMDRequireDependenciesBlockParserPlugin;
