// webpack/lib/dependencies/CommonJsImportsParserPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const { fileURLToPath ***REMOVED*** = require("url");
const CommentCompilationWarning = require("../CommentCompilationWarning");
const RuntimeGlobals = require("../RuntimeGlobals");
const UnsupportedFeatureWarning = require("../UnsupportedFeatureWarning");
const WebpackError = require("../WebpackError");
const BasicEvaluatedExpression = require("../javascript/BasicEvaluatedExpression");
const { VariableInfo ***REMOVED*** = require("../javascript/JavascriptParser");
const {
	evaluateToIdentifier,
	evaluateToString,
	expressionIsUnsupported,
	toConstantDependency
***REMOVED*** = require("../javascript/JavascriptParserHelpers");
const CommonJsFullRequireDependency = require("./CommonJsFullRequireDependency");
const CommonJsRequireContextDependency = require("./CommonJsRequireContextDependency");
const CommonJsRequireDependency = require("./CommonJsRequireDependency");
const ConstDependency = require("./ConstDependency");
const ContextDependencyHelpers = require("./ContextDependencyHelpers");
const LocalModuleDependency = require("./LocalModuleDependency");
const { getLocalModule ***REMOVED*** = require("./LocalModulesHelpers");
const RequireHeaderDependency = require("./RequireHeaderDependency");
const RequireResolveContextDependency = require("./RequireResolveContextDependency");
const RequireResolveDependency = require("./RequireResolveDependency");
const RequireResolveHeaderDependency = require("./RequireResolveHeaderDependency");

/** @typedef {import("estree").CallExpression***REMOVED*** CallExpression */
/** @typedef {import("estree").Expression***REMOVED*** Expression */
/** @typedef {import("estree").NewExpression***REMOVED*** NewExpression */
/** @typedef {import("../../declarations/WebpackOptions").JavascriptParserOptions***REMOVED*** JavascriptParserOptions */
/** @typedef {import("../Dependency").DependencyLocation***REMOVED*** DependencyLocation */
/** @typedef {import("../javascript/JavascriptParser")***REMOVED*** JavascriptParser */
/** @typedef {import("../javascript/JavascriptParser").ImportSource***REMOVED*** ImportSource */
/** @typedef {import("../javascript/JavascriptParser").Range***REMOVED*** Range */

const createRequireSpecifierTag = Symbol("createRequire");
const createdRequireIdentifierTag = Symbol("createRequire()");

class CommonJsImportsParserPlugin {
	/**
	 * @param {JavascriptParserOptions***REMOVED*** options parser options
	 */
	constructor(options) {
		this.options = options;
	***REMOVED***

	/**
	 * @param {JavascriptParser***REMOVED*** parser the parser
	 * @returns {void***REMOVED***
	 */
	apply(parser) {
		const options = this.options;

		const getContext = () => {
			if (parser.currentTagData) {
				const { context ***REMOVED*** = parser.currentTagData;
				return context;
			***REMOVED***
		***REMOVED***;

		// #region metadata
		/**
		 * @param {string***REMOVED*** expression expression
		 * @param {() => string[]***REMOVED*** getMembers get members
		 */
		const tapRequireExpression = (expression, getMembers) => {
			parser.hooks.typeof
				.for(expression)
				.tap(
					"CommonJsImportsParserPlugin",
					toConstantDependency(parser, JSON.stringify("function"))
				);
			parser.hooks.evaluateTypeof
				.for(expression)
				.tap("CommonJsImportsParserPlugin", evaluateToString("function"));
			parser.hooks.evaluateIdentifier
				.for(expression)
				.tap(
					"CommonJsImportsParserPlugin",
					evaluateToIdentifier(expression, "require", getMembers, true)
				);
		***REMOVED***;
		/**
		 * @param {string | symbol***REMOVED*** tag tag
		 */
		const tapRequireExpressionTag = tag => {
			parser.hooks.typeof
				.for(tag)
				.tap(
					"CommonJsImportsParserPlugin",
					toConstantDependency(parser, JSON.stringify("function"))
				);
			parser.hooks.evaluateTypeof
				.for(tag)
				.tap("CommonJsImportsParserPlugin", evaluateToString("function"));
		***REMOVED***;
		tapRequireExpression("require", () => []);
		tapRequireExpression("require.resolve", () => ["resolve"]);
		tapRequireExpression("require.resolveWeak", () => ["resolveWeak"]);
		// #endregion

		// Weird stuff //
		parser.hooks.assign
			.for("require")
			.tap("CommonJsImportsParserPlugin", expr => {
				// to not leak to global "require", we need to define a local require here.
				const dep = new ConstDependency("var require;", 0);
				dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
				parser.state.module.addPresentationalDependency(dep);
				return true;
			***REMOVED***);

		// #region Unsupported
		parser.hooks.expression
			.for("require.main")
			.tap(
				"CommonJsImportsParserPlugin",
				expressionIsUnsupported(
					parser,
					"require.main is not supported by webpack."
				)
			);
		parser.hooks.call
			.for("require.main.require")
			.tap(
				"CommonJsImportsParserPlugin",
				expressionIsUnsupported(
					parser,
					"require.main.require is not supported by webpack."
				)
			);
		parser.hooks.expression
			.for("module.parent.require")
			.tap(
				"CommonJsImportsParserPlugin",
				expressionIsUnsupported(
					parser,
					"module.parent.require is not supported by webpack."
				)
			);
		parser.hooks.call
			.for("module.parent.require")
			.tap(
				"CommonJsImportsParserPlugin",
				expressionIsUnsupported(
					parser,
					"module.parent.require is not supported by webpack."
				)
			);
		// #endregion

		// #region Renaming
		/**
		 * @param {Expression***REMOVED*** expr expression
		 * @returns {boolean***REMOVED*** true when set undefined
		 */
		const defineUndefined = expr => {
			// To avoid "not defined" error, replace the value with undefined
			const dep = new ConstDependency(
				"undefined",
				/** @type {Range***REMOVED*** */ (expr.range)
			);
			dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
			parser.state.module.addPresentationalDependency(dep);
			return false;
		***REMOVED***;
		parser.hooks.canRename
			.for("require")
			.tap("CommonJsImportsParserPlugin", () => true);
		parser.hooks.rename
			.for("require")
			.tap("CommonJsImportsParserPlugin", defineUndefined);
		// #endregion

		// #region Inspection
		const requireCache = toConstantDependency(
			parser,
			RuntimeGlobals.moduleCache,
			[
				RuntimeGlobals.moduleCache,
				RuntimeGlobals.moduleId,
				RuntimeGlobals.moduleLoaded
			]
		);

		parser.hooks.expression
			.for("require.cache")
			.tap("CommonJsImportsParserPlugin", requireCache);
		// #endregion

		// #region Require as expression
		/**
		 * @param {Expression***REMOVED*** expr expression
		 * @returns {boolean***REMOVED*** true when handled
		 */
		const requireAsExpressionHandler = expr => {
			const dep = new CommonJsRequireContextDependency(
				{
					request: /** @type {string***REMOVED*** */ (options.unknownContextRequest),
					recursive: /** @type {boolean***REMOVED*** */ (options.unknownContextRecursive),
					regExp: /** @type {RegExp***REMOVED*** */ (options.unknownContextRegExp),
					mode: "sync"
				***REMOVED***,
				/** @type {Range***REMOVED*** */ (expr.range),
				undefined,
				parser.scope.inShorthand,
				getContext()
			);
			dep.critical =
				options.unknownContextCritical &&
				"require function is used in a way in which dependencies cannot be statically extracted";
			dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
			dep.optional = Boolean(parser.scope.inTry);
			parser.state.current.addDependency(dep);
			return true;
		***REMOVED***;
		parser.hooks.expression
			.for("require")
			.tap("CommonJsImportsParserPlugin", requireAsExpressionHandler);
		// #endregion

		// #region Require
		/**
		 * @param {CallExpression | NewExpression***REMOVED*** expr expression
		 * @param {BasicEvaluatedExpression***REMOVED*** param param
		 * @returns {boolean | void***REMOVED*** true when handled
		 */
		const processRequireItem = (expr, param) => {
			if (param.isString()) {
				const dep = new CommonJsRequireDependency(
					/** @type {string***REMOVED*** */ (param.string),
					/** @type {Range***REMOVED*** */ (param.range),
					getContext()
				);
				dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
				dep.optional = Boolean(parser.scope.inTry);
				parser.state.current.addDependency(dep);
				return true;
			***REMOVED***
		***REMOVED***;
		/**
		 * @param {CallExpression | NewExpression***REMOVED*** expr expression
		 * @param {BasicEvaluatedExpression***REMOVED*** param param
		 * @returns {boolean | void***REMOVED*** true when handled
		 */
		const processRequireContext = (expr, param) => {
			const dep = ContextDependencyHelpers.create(
				CommonJsRequireContextDependency,
				/** @type {Range***REMOVED*** */ (expr.range),
				param,
				expr,
				options,
				{
					category: "commonjs"
				***REMOVED***,
				parser,
				undefined,
				getContext()
			);
			if (!dep) return;
			dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
			dep.optional = Boolean(parser.scope.inTry);
			parser.state.current.addDependency(dep);
			return true;
		***REMOVED***;
		/**
		 * @param {boolean***REMOVED*** callNew true, when require is called with new
		 * @returns {(expr: CallExpression | NewExpression) => (boolean | void)***REMOVED*** handler
		 */
		const createRequireHandler = callNew => expr => {
			if (options.commonjsMagicComments) {
				const { options: requireOptions, errors: commentErrors ***REMOVED*** =
					parser.parseCommentOptions(/** @type {Range***REMOVED*** */ (expr.range));

				if (commentErrors) {
					for (const e of commentErrors) {
						const { comment ***REMOVED*** = e;
						parser.state.module.addWarning(
							new CommentCompilationWarning(
								`Compilation error while processing magic comment(-s): /*${comment.value***REMOVED****/: ${e.message***REMOVED***`,
								/** @type {DependencyLocation***REMOVED*** */ (comment.loc)
							)
						);
					***REMOVED***
				***REMOVED***
				if (requireOptions && requireOptions.webpackIgnore !== undefined) {
					if (typeof requireOptions.webpackIgnore !== "boolean") {
						parser.state.module.addWarning(
							new UnsupportedFeatureWarning(
								`\`webpackIgnore\` expected a boolean, but received: ${requireOptions.webpackIgnore***REMOVED***.`,
								/** @type {DependencyLocation***REMOVED*** */ (expr.loc)
							)
						);
					***REMOVED*** else if (requireOptions.webpackIgnore) {
						// Do not instrument `require()` if `webpackIgnore` is `true`
						return true;
					***REMOVED***
				***REMOVED***
			***REMOVED***

			if (expr.arguments.length !== 1) return;
			let localModule;
			const param = parser.evaluateExpression(expr.arguments[0]);
			if (param.isConditional()) {
				let isExpression = false;
				for (const p of /** @type {BasicEvaluatedExpression[]***REMOVED*** */ (
					param.options
				)) {
					const result = processRequireItem(expr, p);
					if (result === undefined) {
						isExpression = true;
					***REMOVED***
				***REMOVED***
				if (!isExpression) {
					const dep = new RequireHeaderDependency(
						/** @type {Range***REMOVED*** */ (expr.callee.range)
					);
					dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
					parser.state.module.addPresentationalDependency(dep);
					return true;
				***REMOVED***
			***REMOVED***
			if (
				param.isString() &&
				(localModule = getLocalModule(
					parser.state,
					/** @type {string***REMOVED*** */ (param.string)
				))
			) {
				localModule.flagUsed();
				const dep = new LocalModuleDependency(
					localModule,
					/** @type {Range***REMOVED*** */ (expr.range),
					callNew
				);
				dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
				parser.state.module.addPresentationalDependency(dep);
			***REMOVED*** else {
				const result = processRequireItem(expr, param);
				if (result === undefined) {
					processRequireContext(expr, param);
				***REMOVED*** else {
					const dep = new RequireHeaderDependency(
						/** @type {Range***REMOVED*** */ (expr.callee.range)
					);
					dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
					parser.state.module.addPresentationalDependency(dep);
				***REMOVED***
			***REMOVED***
			return true;
		***REMOVED***;
		parser.hooks.call
			.for("require")
			.tap("CommonJsImportsParserPlugin", createRequireHandler(false));
		parser.hooks.new
			.for("require")
			.tap("CommonJsImportsParserPlugin", createRequireHandler(true));
		parser.hooks.call
			.for("module.require")
			.tap("CommonJsImportsParserPlugin", createRequireHandler(false));
		parser.hooks.new
			.for("module.require")
			.tap("CommonJsImportsParserPlugin", createRequireHandler(true));
		// #endregion

		// #region Require with property access
		/**
		 * @param {Expression***REMOVED*** expr expression
		 * @param {string[]***REMOVED*** calleeMembers callee members
		 * @param {CallExpression***REMOVED*** callExpr call expression
		 * @param {string[]***REMOVED*** members members
		 * @param {Range[]***REMOVED*** memberRanges member ranges
		 * @returns {boolean | void***REMOVED*** true when handled
		 */
		const chainHandler = (
			expr,
			calleeMembers,
			callExpr,
			members,
			memberRanges
		) => {
			if (callExpr.arguments.length !== 1) return;
			const param = parser.evaluateExpression(callExpr.arguments[0]);
			if (
				param.isString() &&
				!getLocalModule(parser.state, /** @type {string***REMOVED*** */ (param.string))
			) {
				const dep = new CommonJsFullRequireDependency(
					/** @type {string***REMOVED*** */ (param.string),
					/** @type {Range***REMOVED*** */ (expr.range),
					members,
					/** @type {Range[]***REMOVED*** */ memberRanges
				);
				dep.asiSafe = !parser.isAsiPosition(
					/** @type {Range***REMOVED*** */ (expr.range)[0]
				);
				dep.optional = Boolean(parser.scope.inTry);
				dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
				parser.state.current.addDependency(dep);
				return true;
			***REMOVED***
		***REMOVED***;
		/**
		 * @param {CallExpression***REMOVED*** expr expression
		 * @param {string[]***REMOVED*** calleeMembers callee members
		 * @param {CallExpression***REMOVED*** callExpr call expression
		 * @param {string[]***REMOVED*** members members
		 * @param {Range[]***REMOVED*** memberRanges member ranges
		 * @returns {boolean | void***REMOVED*** true when handled
		 */
		const callChainHandler = (
			expr,
			calleeMembers,
			callExpr,
			members,
			memberRanges
		) => {
			if (callExpr.arguments.length !== 1) return;
			const param = parser.evaluateExpression(callExpr.arguments[0]);
			if (
				param.isString() &&
				!getLocalModule(parser.state, /** @type {string***REMOVED*** */ (param.string))
			) {
				const dep = new CommonJsFullRequireDependency(
					/** @type {string***REMOVED*** */ (param.string),
					/** @type {Range***REMOVED*** */ (expr.callee.range),
					members,
					/** @type {Range[]***REMOVED*** */ memberRanges
				);
				dep.call = true;
				dep.asiSafe = !parser.isAsiPosition(
					/** @type {Range***REMOVED*** */ (expr.range)[0]
				);
				dep.optional = Boolean(parser.scope.inTry);
				dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.callee.loc);
				parser.state.current.addDependency(dep);
				parser.walkExpressions(expr.arguments);
				return true;
			***REMOVED***
		***REMOVED***;
		parser.hooks.memberChainOfCallMemberChain
			.for("require")
			.tap("CommonJsImportsParserPlugin", chainHandler);
		parser.hooks.memberChainOfCallMemberChain
			.for("module.require")
			.tap("CommonJsImportsParserPlugin", chainHandler);
		parser.hooks.callMemberChainOfCallMemberChain
			.for("require")
			.tap("CommonJsImportsParserPlugin", callChainHandler);
		parser.hooks.callMemberChainOfCallMemberChain
			.for("module.require")
			.tap("CommonJsImportsParserPlugin", callChainHandler);
		// #endregion

		// #region Require.resolve
		/**
		 * @param {CallExpression***REMOVED*** expr call expression
		 * @param {boolean***REMOVED*** weak weak
		 * @returns {boolean | void***REMOVED*** true when handled
		 */
		const processResolve = (expr, weak) => {
			if (!weak && options.commonjsMagicComments) {
				const { options: requireOptions, errors: commentErrors ***REMOVED*** =
					parser.parseCommentOptions(/** @type {Range***REMOVED*** */ (expr.range));

				if (commentErrors) {
					for (const e of commentErrors) {
						const { comment ***REMOVED*** = e;
						parser.state.module.addWarning(
							new CommentCompilationWarning(
								`Compilation error while processing magic comment(-s): /*${comment.value***REMOVED****/: ${e.message***REMOVED***`,
								/** @type {DependencyLocation***REMOVED*** */ (comment.loc)
							)
						);
					***REMOVED***
				***REMOVED***
				if (requireOptions && requireOptions.webpackIgnore !== undefined) {
					if (typeof requireOptions.webpackIgnore !== "boolean") {
						parser.state.module.addWarning(
							new UnsupportedFeatureWarning(
								`\`webpackIgnore\` expected a boolean, but received: ${requireOptions.webpackIgnore***REMOVED***.`,
								/** @type {DependencyLocation***REMOVED*** */ (expr.loc)
							)
						);
					***REMOVED*** else if (requireOptions.webpackIgnore) {
						// Do not instrument `require()` if `webpackIgnore` is `true`
						return true;
					***REMOVED***
				***REMOVED***
			***REMOVED***

			if (expr.arguments.length !== 1) return;
			const param = parser.evaluateExpression(expr.arguments[0]);
			if (param.isConditional()) {
				for (const option of /** @type {BasicEvaluatedExpression[]***REMOVED*** */ (
					param.options
				)) {
					const result = processResolveItem(expr, option, weak);
					if (result === undefined) {
						processResolveContext(expr, option, weak);
					***REMOVED***
				***REMOVED***
				const dep = new RequireResolveHeaderDependency(
					/** @type {Range***REMOVED*** */ (expr.callee.range)
				);
				dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
				parser.state.module.addPresentationalDependency(dep);
				return true;
			***REMOVED***
			const result = processResolveItem(expr, param, weak);
			if (result === undefined) {
				processResolveContext(expr, param, weak);
			***REMOVED***
			const dep = new RequireResolveHeaderDependency(
				/** @type {Range***REMOVED*** */ (expr.callee.range)
			);
			dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
			parser.state.module.addPresentationalDependency(dep);
			return true;
		***REMOVED***;
		/**
		 * @param {CallExpression***REMOVED*** expr call expression
		 * @param {BasicEvaluatedExpression***REMOVED*** param param
		 * @param {boolean***REMOVED*** weak weak
		 * @returns {boolean | void***REMOVED*** true when handled
		 */
		const processResolveItem = (expr, param, weak) => {
			if (param.isString()) {
				const dep = new RequireResolveDependency(
					/** @type {string***REMOVED*** */ (param.string),
					/** @type {Range***REMOVED*** */ (param.range),
					getContext()
				);
				dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
				dep.optional = Boolean(parser.scope.inTry);
				dep.weak = weak;
				parser.state.current.addDependency(dep);
				return true;
			***REMOVED***
		***REMOVED***;
		/**
		 * @param {CallExpression***REMOVED*** expr call expression
		 * @param {BasicEvaluatedExpression***REMOVED*** param param
		 * @param {boolean***REMOVED*** weak weak
		 * @returns {boolean | void***REMOVED*** true when handled
		 */
		const processResolveContext = (expr, param, weak) => {
			const dep = ContextDependencyHelpers.create(
				RequireResolveContextDependency,
				/** @type {Range***REMOVED*** */ (param.range),
				param,
				expr,
				options,
				{
					category: "commonjs",
					mode: weak ? "weak" : "sync"
				***REMOVED***,
				parser,
				getContext()
			);
			if (!dep) return;
			dep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
			dep.optional = Boolean(parser.scope.inTry);
			parser.state.current.addDependency(dep);
			return true;
		***REMOVED***;

		parser.hooks.call
			.for("require.resolve")
			.tap("CommonJsImportsParserPlugin", expr => processResolve(expr, false));
		parser.hooks.call
			.for("require.resolveWeak")
			.tap("CommonJsImportsParserPlugin", expr => processResolve(expr, true));
		// #endregion

		// #region Create require

		if (!options.createRequire) return;

		/** @type {ImportSource[]***REMOVED*** */
		let moduleName = [];
		/** @type {string | undefined***REMOVED*** */
		let specifierName;

		if (options.createRequire === true) {
			moduleName = ["module", "node:module"];
			specifierName = "createRequire";
		***REMOVED*** else {
			let moduleName;
			const match = /^(.*) from (.*)$/.exec(options.createRequire);
			if (match) {
				[, specifierName, moduleName] = match;
			***REMOVED***
			if (!specifierName || !moduleName) {
				const err = new WebpackError(
					`Parsing javascript parser option "createRequire" failed, got ${JSON.stringify(
						options.createRequire
					)***REMOVED***`
				);
				err.details =
					'Expected string in format "createRequire from module", where "createRequire" is specifier name and "module" name of the module';
				throw err;
			***REMOVED***
		***REMOVED***

		tapRequireExpressionTag(createdRequireIdentifierTag);
		tapRequireExpressionTag(createRequireSpecifierTag);
		parser.hooks.evaluateCallExpression
			.for(createRequireSpecifierTag)
			.tap("CommonJsImportsParserPlugin", expr => {
				const context = parseCreateRequireArguments(expr);
				if (context === undefined) return;
				const ident = parser.evaluatedVariable({
					tag: createdRequireIdentifierTag,
					data: { context ***REMOVED***,
					next: undefined
				***REMOVED***);

				return new BasicEvaluatedExpression()
					.setIdentifier(ident, ident, () => [])
					.setSideEffects(false)
					.setRange(/** @type {Range***REMOVED*** */ (expr.range));
			***REMOVED***);
		parser.hooks.unhandledExpressionMemberChain
			.for(createdRequireIdentifierTag)
			.tap("CommonJsImportsParserPlugin", (expr, members) =>
				expressionIsUnsupported(
					parser,
					`createRequire().${members.join(".")***REMOVED*** is not supported by webpack.`
				)(expr)
			);
		parser.hooks.canRename
			.for(createdRequireIdentifierTag)
			.tap("CommonJsImportsParserPlugin", () => true);
		parser.hooks.canRename
			.for(createRequireSpecifierTag)
			.tap("CommonJsImportsParserPlugin", () => true);
		parser.hooks.rename
			.for(createRequireSpecifierTag)
			.tap("CommonJsImportsParserPlugin", defineUndefined);
		parser.hooks.expression
			.for(createdRequireIdentifierTag)
			.tap("CommonJsImportsParserPlugin", requireAsExpressionHandler);
		parser.hooks.call
			.for(createdRequireIdentifierTag)
			.tap("CommonJsImportsParserPlugin", createRequireHandler(false));
		/**
		 * @param {CallExpression***REMOVED*** expr call expression
		 * @returns {string | void***REMOVED*** context
		 */
		const parseCreateRequireArguments = expr => {
			const args = expr.arguments;
			if (args.length !== 1) {
				const err = new WebpackError(
					"module.createRequire supports only one argument."
				);
				err.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
				parser.state.module.addWarning(err);
				return;
			***REMOVED***
			const arg = args[0];
			const evaluated = parser.evaluateExpression(arg);
			if (!evaluated.isString()) {
				const err = new WebpackError(
					"module.createRequire failed parsing argument."
				);
				err.loc = /** @type {DependencyLocation***REMOVED*** */ (arg.loc);
				parser.state.module.addWarning(err);
				return;
			***REMOVED***
			const ctx = /** @type {string***REMOVED*** */ (evaluated.string).startsWith("file://")
				? fileURLToPath(/** @type {string***REMOVED*** */ (evaluated.string))
				: /** @type {string***REMOVED*** */ (evaluated.string);
			// argument always should be a filename
			return ctx.slice(0, ctx.lastIndexOf(ctx.startsWith("/") ? "/" : "\\"));
		***REMOVED***;

		parser.hooks.import.tap(
			{
				name: "CommonJsImportsParserPlugin",
				stage: -10
			***REMOVED***,
			(statement, source) => {
				if (
					!moduleName.includes(source) ||
					statement.specifiers.length !== 1 ||
					statement.specifiers[0].type !== "ImportSpecifier" ||
					statement.specifiers[0].imported.type !== "Identifier" ||
					statement.specifiers[0].imported.name !== specifierName
				)
					return;
				// clear for 'import { createRequire as x ***REMOVED*** from "module"'
				// if any other specifier was used import module
				const clearDep = new ConstDependency(
					parser.isAsiPosition(/** @type {Range***REMOVED*** */ (statement.range)[0])
						? ";"
						: "",
					/** @type {Range***REMOVED*** */ (statement.range)
				);
				clearDep.loc = /** @type {DependencyLocation***REMOVED*** */ (statement.loc);
				parser.state.module.addPresentationalDependency(clearDep);
				parser.unsetAsiPosition(/** @type {Range***REMOVED*** */ (statement.range)[1]);
				return true;
			***REMOVED***
		);
		parser.hooks.importSpecifier.tap(
			{
				name: "CommonJsImportsParserPlugin",
				stage: -10
			***REMOVED***,
			(statement, source, id, name) => {
				if (!moduleName.includes(source) || id !== specifierName) return;
				parser.tagVariable(name, createRequireSpecifierTag);
				return true;
			***REMOVED***
		);
		parser.hooks.preDeclarator.tap(
			"CommonJsImportsParserPlugin",
			declarator => {
				if (
					declarator.id.type !== "Identifier" ||
					!declarator.init ||
					declarator.init.type !== "CallExpression" ||
					declarator.init.callee.type !== "Identifier"
				)
					return;
				const variableInfo = parser.getVariableInfo(
					declarator.init.callee.name
				);
				if (
					variableInfo instanceof VariableInfo &&
					variableInfo.tagInfo &&
					variableInfo.tagInfo.tag === createRequireSpecifierTag
				) {
					const context = parseCreateRequireArguments(declarator.init);
					if (context === undefined) return;
					parser.tagVariable(declarator.id.name, createdRequireIdentifierTag, {
						name: declarator.id.name,
						context
					***REMOVED***);
					return true;
				***REMOVED***
			***REMOVED***
		);

		parser.hooks.memberChainOfCallMemberChain
			.for(createRequireSpecifierTag)
			.tap(
				"CommonJsImportsParserPlugin",
				(expr, calleeMembers, callExpr, members) => {
					if (
						calleeMembers.length !== 0 ||
						members.length !== 1 ||
						members[0] !== "cache"
					)
						return;
					// createRequire().cache
					const context = parseCreateRequireArguments(callExpr);
					if (context === undefined) return;
					return requireCache(expr);
				***REMOVED***
			);
		parser.hooks.callMemberChainOfCallMemberChain
			.for(createRequireSpecifierTag)
			.tap(
				"CommonJsImportsParserPlugin",
				(expr, calleeMembers, innerCallExpression, members) => {
					if (
						calleeMembers.length !== 0 ||
						members.length !== 1 ||
						members[0] !== "resolve"
					)
						return;
					// createRequire().resolve()
					return processResolve(expr, false);
				***REMOVED***
			);
		parser.hooks.expressionMemberChain
			.for(createdRequireIdentifierTag)
			.tap("CommonJsImportsParserPlugin", (expr, members) => {
				// require.cache
				if (members.length === 1 && members[0] === "cache") {
					return requireCache(expr);
				***REMOVED***
			***REMOVED***);
		parser.hooks.callMemberChain
			.for(createdRequireIdentifierTag)
			.tap("CommonJsImportsParserPlugin", (expr, members) => {
				// require.resolve()
				if (members.length === 1 && members[0] === "resolve") {
					return processResolve(expr, false);
				***REMOVED***
			***REMOVED***);
		parser.hooks.call
			.for(createRequireSpecifierTag)
			.tap("CommonJsImportsParserPlugin", expr => {
				const clearDep = new ConstDependency(
					"/* createRequire() */ undefined",
					/** @type {Range***REMOVED*** */ (expr.range)
				);
				clearDep.loc = /** @type {DependencyLocation***REMOVED*** */ (expr.loc);
				parser.state.module.addPresentationalDependency(clearDep);
				return true;
			***REMOVED***);
		// #endregion
	***REMOVED***
***REMOVED***
module.exports = CommonJsImportsParserPlugin;
