// webpack/lib/dependencies/CssIcssImportDependency.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Ivan Kopeykin @vankop
*/

"use strict";

const makeSerializable = require("../util/makeSerializable");
const CssIcssExportDependency = require("./CssIcssExportDependency");
const CssLocalIdentifierDependency = require("./CssLocalIdentifierDependency");
const ModuleDependency = require("./ModuleDependency");

/** @typedef {import("webpack-sources").ReplaceSource***REMOVED*** ReplaceSource */
/** @typedef {import("../Dependency")***REMOVED*** Dependency */
/** @typedef {import("../Dependency").ExportsSpec***REMOVED*** ExportsSpec */
/** @typedef {import("../DependencyTemplate").CssDependencyTemplateContext***REMOVED*** DependencyTemplateContext */
/** @typedef {import("../Module")***REMOVED*** Module */
/** @typedef {import("../ModuleGraph")***REMOVED*** ModuleGraph */
/** @typedef {import("../javascript/JavascriptParser").Range***REMOVED*** Range */
/** @typedef {import("../serialization/ObjectMiddleware").ObjectDeserializerContext***REMOVED*** ObjectDeserializerContext */
/** @typedef {import("../serialization/ObjectMiddleware").ObjectSerializerContext***REMOVED*** ObjectSerializerContext */

class CssIcssImportDependency extends ModuleDependency {
	/**
	 * Example of dependency:
	 *
	 *:import('./style.css') { IMPORTED_NAME: v-primary ***REMOVED***
	 * @param {string***REMOVED*** request request request path which needs resolving
	 * @param {string***REMOVED*** exportName export name
	 * @param {Range***REMOVED*** range the range of dependency
	 */
	constructor(request, exportName, range) {
		super(request);
		this.exportName = exportName;
		this.range = range;
	***REMOVED***

	get type() {
		return "css :import";
	***REMOVED***

	get category() {
		return "css-import";
	***REMOVED***

	/**
	 * @param {ObjectSerializerContext***REMOVED*** context context
	 */
	serialize(context) {
		const { write ***REMOVED*** = context;
		write(this.range);
		write(this.exportName);
		super.serialize(context);
	***REMOVED***

	/**
	 * @param {ObjectDeserializerContext***REMOVED*** context context
	 */
	deserialize(context) {
		const { read ***REMOVED*** = context;
		this.range = read();
		this.exportName = read();
		super.deserialize(context);
	***REMOVED***
***REMOVED***

CssIcssImportDependency.Template = class CssIcssImportDependencyTemplate extends (
	ModuleDependency.Template
) {
	/**
	 * @param {Dependency***REMOVED*** dependency the dependency for which the template should be applied
	 * @param {ReplaceSource***REMOVED*** source the current replace source which can be modified
	 * @param {DependencyTemplateContext***REMOVED*** templateContext the context object
	 * @returns {void***REMOVED***
	 */
	apply(dependency, source, templateContext) {
		const dep = /** @type {CssIcssImportDependency***REMOVED*** */ (dependency);
		const { range ***REMOVED*** = dep;
		const module =
			/** @type {Module***REMOVED*** */
			(templateContext.moduleGraph.getModule(dep));
		let value;

		for (const item of module.dependencies) {
			if (
				item instanceof CssLocalIdentifierDependency &&
				dep.exportName === item.name
			) {
				value = CssLocalIdentifierDependency.Template.getIdentifier(
					item,
					dep.exportName,
					{
						...templateContext,
						module
					***REMOVED***
				);
				break;
			***REMOVED*** else if (
				item instanceof CssIcssExportDependency &&
				dep.exportName === item.name
			) {
				value = item.value;
				break;
			***REMOVED***
		***REMOVED***

		if (!value) {
			throw new Error(
				`Imported '${dep.exportName***REMOVED***' name from '${dep.request***REMOVED***' not found`
			);
		***REMOVED***

		source.replace(range[0], range[1], value);
	***REMOVED***
***REMOVED***;

makeSerializable(
	CssIcssImportDependency,
	"webpack/lib/dependencies/CssIcssImportDependency"
);

module.exports = CssIcssImportDependency;
