// webpack/lib/dependencies/CssLocalIdentifierDependency.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Ivan Kopeykin @vankop
*/

"use strict";

const { cssExportConvention ***REMOVED*** = require("../util/conventions");
const createHash = require("../util/createHash");
const { makePathsRelative ***REMOVED*** = require("../util/identifier");
const makeSerializable = require("../util/makeSerializable");
const memoize = require("../util/memoize");
const NullDependency = require("./NullDependency");

/** @typedef {import("webpack-sources").ReplaceSource***REMOVED*** ReplaceSource */
/** @typedef {import("../../declarations/WebpackOptions").CssGeneratorExportsConvention***REMOVED*** CssGeneratorExportsConvention */
/** @typedef {import("../../declarations/WebpackOptions").CssGeneratorLocalIdentName***REMOVED*** CssGeneratorLocalIdentName */
/** @typedef {import("../../declarations/WebpackOptions").HashFunction***REMOVED*** HashFunction */
/** @typedef {import("../ChunkGraph")***REMOVED*** ChunkGraph */
/** @typedef {import("../CssModule")***REMOVED*** CssModule */
/** @typedef {import("../Dependency")***REMOVED*** Dependency */
/** @typedef {import("../Dependency").ExportsSpec***REMOVED*** ExportsSpec */
/** @typedef {import("../Dependency").UpdateHashContext***REMOVED*** UpdateHashContext */
/** @typedef {import("../DependencyTemplate").CssDependencyTemplateContext***REMOVED*** DependencyTemplateContext */
/** @typedef {import("../ModuleGraph")***REMOVED*** ModuleGraph */
/** @typedef {import("../NormalModuleFactory").ResourceDataWithData***REMOVED*** ResourceDataWithData */
/** @typedef {import("../RuntimeTemplate")***REMOVED*** RuntimeTemplate */
/** @typedef {import("../css/CssGenerator")***REMOVED*** CssGenerator */
/** @typedef {import("../css/CssParser").Range***REMOVED*** Range */
/** @typedef {import("../serialization/ObjectMiddleware").ObjectDeserializerContext***REMOVED*** ObjectDeserializerContext */
/** @typedef {import("../serialization/ObjectMiddleware").ObjectSerializerContext***REMOVED*** ObjectSerializerContext */
/** @typedef {import("../util/Hash")***REMOVED*** Hash */

const getCssParser = memoize(() => require("../css/CssParser"));

/**
 * @param {string***REMOVED*** local css local
 * @param {CssModule***REMOVED*** module module
 * @param {ChunkGraph***REMOVED*** chunkGraph chunk graph
 * @param {RuntimeTemplate***REMOVED*** runtimeTemplate runtime template
 * @returns {string***REMOVED*** local ident
 */
const getLocalIdent = (local, module, chunkGraph, runtimeTemplate) => {
	const generator = /** @type {CssGenerator***REMOVED*** */ (module.generator);
	const localIdentName =
		/** @type {CssGeneratorLocalIdentName***REMOVED*** */
		(generator.localIdentName);
	const relativeResourcePath = makePathsRelative(
		/** @type {string***REMOVED*** */
		(module.context),
		module.matchResource || module.resource,
		runtimeTemplate.compilation.compiler.root
	);
	const { hashFunction, hashDigest, hashDigestLength, hashSalt, uniqueName ***REMOVED*** =
		runtimeTemplate.outputOptions;
	const hash = createHash(/** @type {HashFunction***REMOVED*** */ (hashFunction));

	if (hashSalt) {
		hash.update(hashSalt);
	***REMOVED***

	hash.update(relativeResourcePath);

	if (!/\[local\]/.test(localIdentName)) {
		hash.update(local);
	***REMOVED***

	const localIdentHash =
		/** @type {string***REMOVED*** */
		(hash.digest(hashDigest)).slice(0, hashDigestLength);

	return runtimeTemplate.compilation
		.getPath(localIdentName, {
			filename: relativeResourcePath,
			hash: localIdentHash,
			contentHash: localIdentHash,
			chunkGraph,
			module
		***REMOVED***)
		.replace(/\[local\]/g, local)
		.replace(/\[uniqueName\]/g, /** @type {string***REMOVED*** */ (uniqueName))
		.replace(/^((-?[0-9])|--)/, "_$1");
***REMOVED***;

class CssLocalIdentifierDependency extends NullDependency {
	/**
	 * @param {string***REMOVED*** name name
	 * @param {Range***REMOVED*** range range
	 * @param {string=***REMOVED*** prefix prefix
	 */
	constructor(name, range, prefix = "") {
		super();
		this.name = name;
		this.range = range;
		this.prefix = prefix;
		this._conventionNames = undefined;
		this._hashUpdate = undefined;
	***REMOVED***

	get type() {
		return "css local identifier";
	***REMOVED***

	/**
	 * @param {string***REMOVED*** name export name
	 * @param {CssGeneratorExportsConvention***REMOVED*** convention convention of the export name
	 * @returns {string[]***REMOVED*** convention results
	 */
	getExportsConventionNames(name, convention) {
		if (this._conventionNames) {
			return this._conventionNames;
		***REMOVED***
		this._conventionNames = cssExportConvention(this.name, convention);
		return this._conventionNames;
	***REMOVED***

	/**
	 * Returns the exported names
	 * @param {ModuleGraph***REMOVED*** moduleGraph module graph
	 * @returns {ExportsSpec | undefined***REMOVED*** export names
	 */
	getExports(moduleGraph) {
		const module = /** @type {CssModule***REMOVED*** */ (moduleGraph.getParentModule(this));
		const generator = /** @type {CssGenerator***REMOVED*** */ (module.generator);
		const names = this.getExportsConventionNames(
			this.name,
			/** @type {CssGeneratorExportsConvention***REMOVED*** */ (generator.convention)
		);
		return {
			exports: names.map(name => ({
				name,
				canMangle: true
			***REMOVED***)),
			dependencies: undefined
		***REMOVED***;
	***REMOVED***

	/**
	 * Update the hash
	 * @param {Hash***REMOVED*** hash hash to be updated
	 * @param {UpdateHashContext***REMOVED*** context context
	 * @returns {void***REMOVED***
	 */
	updateHash(hash, { chunkGraph ***REMOVED***) {
		if (this._hashUpdate === undefined) {
			const module =
				/** @type {CssModule***REMOVED*** */
				(chunkGraph.moduleGraph.getParentModule(this));
			const generator = /** @type {CssGenerator***REMOVED*** */ (module.generator);
			const names = this.getExportsConventionNames(
				this.name,
				/** @type {CssGeneratorExportsConvention***REMOVED*** */
				(generator.convention)
			);
			this._hashUpdate = `exportsConvention|${JSON.stringify(names)***REMOVED***|localIdentName|${JSON.stringify(generator.localIdentName)***REMOVED***`;
		***REMOVED***
		hash.update(this._hashUpdate);
	***REMOVED***

	/**
	 * @param {ObjectSerializerContext***REMOVED*** context context
	 */
	serialize(context) {
		const { write ***REMOVED*** = context;
		write(this.name);
		write(this.range);
		write(this.prefix);
		super.serialize(context);
	***REMOVED***

	/**
	 * @param {ObjectDeserializerContext***REMOVED*** context context
	 */
	deserialize(context) {
		const { read ***REMOVED*** = context;
		this.name = read();
		this.range = read();
		this.prefix = read();
		super.deserialize(context);
	***REMOVED***
***REMOVED***

CssLocalIdentifierDependency.Template = class CssLocalIdentifierDependencyTemplate extends (
	NullDependency.Template
) {
	/**
	 * @param {Dependency***REMOVED*** dependency the dependency for which the template should be applied
	 * @param {string***REMOVED*** local local name
	 * @param {DependencyTemplateContext***REMOVED*** templateContext the context object
	 * @returns {string***REMOVED*** identifier
	 */
	static getIdentifier(
		dependency,
		local,
		{ module: m, chunkGraph, runtimeTemplate ***REMOVED***
	) {
		const dep = /** @type {CssLocalIdentifierDependency***REMOVED*** */ (dependency);
		const module = /** @type {CssModule***REMOVED*** */ (m);

		return (
			dep.prefix +
			getCssParser().escapeIdentifier(
				getLocalIdent(local, module, chunkGraph, runtimeTemplate)
			)
		);
	***REMOVED***

	/**
	 * @param {Dependency***REMOVED*** dependency the dependency for which the template should be applied
	 * @param {ReplaceSource***REMOVED*** source the current replace source which can be modified
	 * @param {DependencyTemplateContext***REMOVED*** templateContext the context object
	 * @returns {void***REMOVED***
	 */
	apply(dependency, source, templateContext) {
		const { module: m, moduleGraph, runtime, cssData ***REMOVED*** = templateContext;
		const dep = /** @type {CssLocalIdentifierDependency***REMOVED*** */ (dependency);
		const module = /** @type {CssModule***REMOVED*** */ (m);
		const generator = /** @type {CssGenerator***REMOVED*** */ (module.generator);
		const names = dep.getExportsConventionNames(
			dep.name,
			/** @type {CssGeneratorExportsConvention***REMOVED*** */
			(generator.convention)
		);
		const usedNames =
			/** @type {(string)[]***REMOVED*** */
			(
				names
					.map(name =>
						moduleGraph.getExportInfo(module, name).getUsedName(name, runtime)
					)
					.filter(Boolean)
			);
		const local = usedNames.length === 0 ? names[0] : usedNames[0];
		const identifier = CssLocalIdentifierDependencyTemplate.getIdentifier(
			dep,
			local,
			templateContext
		);

		source.replace(dep.range[0], dep.range[1] - 1, identifier);

		for (const used of usedNames.concat(names)) {
			cssData.exports.set(used, getCssParser().unescapeIdentifier(identifier));
		***REMOVED***
	***REMOVED***
***REMOVED***;

makeSerializable(
	CssLocalIdentifierDependency,
	"webpack/lib/dependencies/CssLocalIdentifierDependency"
);

module.exports = CssLocalIdentifierDependency;
