// webpack/lib/dependencies/HarmonyEvaluatedImportSpecifierDependency.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Ivan Kopeykin @vankop
*/

"use strict";

const makeSerializable = require("../util/makeSerializable");
const HarmonyImportSpecifierDependency = require("./HarmonyImportSpecifierDependency");

/** @typedef {import("webpack-sources").ReplaceSource***REMOVED*** ReplaceSource */
/** @typedef {import("../ChunkGraph")***REMOVED*** ChunkGraph */
/** @typedef {import("../Dependency")***REMOVED*** Dependency */
/** @typedef {import("../DependencyTemplate").DependencyTemplateContext***REMOVED*** DependencyTemplateContext */
/** @typedef {import("../Module").BuildMeta***REMOVED*** BuildMeta */
/** @typedef {import("../ModuleGraphConnection")***REMOVED*** ModuleGraphConnection */
/** @typedef {import("../javascript/JavascriptParser").ImportAttributes***REMOVED*** ImportAttributes */
/** @typedef {import("../javascript/JavascriptParser").Range***REMOVED*** Range */
/** @typedef {import("../serialization/ObjectMiddleware").ObjectDeserializerContext***REMOVED*** ObjectDeserializerContext */
/** @typedef {import("../serialization/ObjectMiddleware").ObjectSerializerContext***REMOVED*** ObjectSerializerContext */

/**
 * Dependency for static evaluating import specifier. e.g.
 * @example
 * import a from "a";
 * "x" in a;
 * a.x !== undefined; // if x value statically analyzable
 */
class HarmonyEvaluatedImportSpecifierDependency extends HarmonyImportSpecifierDependency {
	/**
	 * @param {string***REMOVED*** request the request string
	 * @param {number***REMOVED*** sourceOrder source order
	 * @param {string[]***REMOVED*** ids ids
	 * @param {string***REMOVED*** name name
	 * @param {Range***REMOVED*** range location in source code
	 * @param {ImportAttributes***REMOVED*** attributes import assertions
	 * @param {string***REMOVED*** operator operator
	 */
	constructor(request, sourceOrder, ids, name, range, attributes, operator) {
		super(request, sourceOrder, ids, name, range, false, attributes, []);
		this.operator = operator;
	***REMOVED***

	get type() {
		return `evaluated X ${this.operator***REMOVED*** harmony import specifier`;
	***REMOVED***

	/**
	 * @param {ObjectSerializerContext***REMOVED*** context context
	 */
	serialize(context) {
		super.serialize(context);
		const { write ***REMOVED*** = context;
		write(this.operator);
	***REMOVED***

	/**
	 * @param {ObjectDeserializerContext***REMOVED*** context context
	 */
	deserialize(context) {
		super.deserialize(context);
		const { read ***REMOVED*** = context;
		this.operator = read();
	***REMOVED***
***REMOVED***

makeSerializable(
	HarmonyEvaluatedImportSpecifierDependency,
	"webpack/lib/dependencies/HarmonyEvaluatedImportSpecifierDependency"
);

HarmonyEvaluatedImportSpecifierDependency.Template = class HarmonyEvaluatedImportSpecifierDependencyTemplate extends (
	HarmonyImportSpecifierDependency.Template
) {
	/**
	 * @param {Dependency***REMOVED*** dependency the dependency for which the template should be applied
	 * @param {ReplaceSource***REMOVED*** source the current replace source which can be modified
	 * @param {DependencyTemplateContext***REMOVED*** templateContext the context object
	 * @returns {void***REMOVED***
	 */
	apply(dependency, source, templateContext) {
		const dep =
			/** @type {HarmonyEvaluatedImportSpecifierDependency***REMOVED*** */
			(dependency);
		const { module, moduleGraph, runtime ***REMOVED*** = templateContext;
		const connection = moduleGraph.getConnection(dep);
		// Skip rendering depending when dependency is conditional
		if (connection && !connection.isTargetActive(runtime)) return;

		const exportsInfo = moduleGraph.getExportsInfo(
			/** @type {ModuleGraphConnection***REMOVED*** */ (connection).module
		);
		const ids = dep.getIds(moduleGraph);

		let value;

		const exportsType =
			/** @type {ModuleGraphConnection***REMOVED*** */
			(connection).module.getExportsType(
				moduleGraph,
				/** @type {BuildMeta***REMOVED*** */
				(module.buildMeta).strictHarmonyModule
			);
		switch (exportsType) {
			case "default-with-named": {
				if (ids[0] === "default") {
					value =
						ids.length === 1 || exportsInfo.isExportProvided(ids.slice(1));
				***REMOVED*** else {
					value = exportsInfo.isExportProvided(ids);
				***REMOVED***
				break;
			***REMOVED***
			case "namespace": {
				value =
					ids[0] === "__esModule"
						? ids.length === 1 || undefined
						: exportsInfo.isExportProvided(ids);
				break;
			***REMOVED***
			case "dynamic": {
				if (ids[0] !== "default") {
					value = exportsInfo.isExportProvided(ids);
				***REMOVED***
				break;
			***REMOVED***
			// default-only could lead to runtime error, when default value is primitive
		***REMOVED***

		if (typeof value === "boolean") {
			source.replace(dep.range[0], dep.range[1] - 1, ` ${value***REMOVED***`);
		***REMOVED*** else {
			const usedName = exportsInfo.getUsedName(ids, runtime);

			const code = this._getCodeForIds(
				dep,
				source,
				templateContext,
				ids.slice(0, -1)
			);
			source.replace(
				dep.range[0],
				dep.range[1] - 1,
				`${
					usedName ? JSON.stringify(usedName[usedName.length - 1]) : '""'
				***REMOVED*** in ${code***REMOVED***`
			);
		***REMOVED***
	***REMOVED***
***REMOVED***;

module.exports = HarmonyEvaluatedImportSpecifierDependency;
