// webpack/lib/dependencies/ImportWeakDependency.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const makeSerializable = require("../util/makeSerializable");
const ImportDependency = require("./ImportDependency");

/** @typedef {import("webpack-sources").ReplaceSource***REMOVED*** ReplaceSource */
/** @typedef {import("../Dependency")***REMOVED*** Dependency */
/** @typedef {import("../Dependency").ReferencedExport***REMOVED*** ReferencedExport */
/** @typedef {import("../DependencyTemplate").DependencyTemplateContext***REMOVED*** DependencyTemplateContext */
/** @typedef {import("../Module")***REMOVED*** Module */
/** @typedef {import("../Module").BuildMeta***REMOVED*** BuildMeta */
/** @typedef {import("../ModuleGraph")***REMOVED*** ModuleGraph */
/** @typedef {import("../javascript/JavascriptParser").ImportAttributes***REMOVED*** ImportAttributes */
/** @typedef {import("../javascript/JavascriptParser").Range***REMOVED*** Range */

class ImportWeakDependency extends ImportDependency {
	/**
	 * @param {string***REMOVED*** request the request
	 * @param {Range***REMOVED*** range expression range
	 * @param {(string[][] | null)=***REMOVED*** referencedExports list of referenced exports
	 * @param {ImportAttributes=***REMOVED*** attributes import attributes
	 */
	constructor(request, range, referencedExports, attributes) {
		super(request, range, referencedExports, attributes);
		this.weak = true;
	***REMOVED***

	get type() {
		return "import() weak";
	***REMOVED***
***REMOVED***

makeSerializable(
	ImportWeakDependency,
	"webpack/lib/dependencies/ImportWeakDependency"
);

ImportWeakDependency.Template = class ImportDependencyTemplate extends (
	ImportDependency.Template
) {
	/**
	 * @param {Dependency***REMOVED*** dependency the dependency for which the template should be applied
	 * @param {ReplaceSource***REMOVED*** source the current replace source which can be modified
	 * @param {DependencyTemplateContext***REMOVED*** templateContext the context object
	 * @returns {void***REMOVED***
	 */
	apply(
		dependency,
		source,
		{ runtimeTemplate, module, moduleGraph, chunkGraph, runtimeRequirements ***REMOVED***
	) {
		const dep = /** @type {ImportWeakDependency***REMOVED*** */ (dependency);
		const content = runtimeTemplate.moduleNamespacePromise({
			chunkGraph,
			module: /** @type {Module***REMOVED*** */ (moduleGraph.getModule(dep)),
			request: dep.request,
			strict: /** @type {BuildMeta***REMOVED*** */ (module.buildMeta).strictHarmonyModule,
			message: "import() weak",
			weak: true,
			runtimeRequirements
		***REMOVED***);

		source.replace(dep.range[0], dep.range[1] - 1, content);
	***REMOVED***
***REMOVED***;

module.exports = ImportWeakDependency;
