// webpack/lib/dependencies/LoaderImportDependency.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const ModuleDependency = require("./ModuleDependency");

/** @typedef {import("../Dependency").GetConditionFn***REMOVED*** GetConditionFn */
/** @typedef {import("../ModuleGraph")***REMOVED*** ModuleGraph */
/** @typedef {import("../ModuleGraphConnection")***REMOVED*** ModuleGraphConnection */
/** @typedef {import("../ModuleGraphConnection").ConnectionState***REMOVED*** ConnectionState */
/** @typedef {import("../util/runtime").RuntimeSpec***REMOVED*** RuntimeSpec */

class LoaderImportDependency extends ModuleDependency {
	/**
	 * @param {string***REMOVED*** request request string
	 */
	constructor(request) {
		super(request);
		this.weak = true;
	***REMOVED***

	get type() {
		return "loader import";
	***REMOVED***

	get category() {
		return "loaderImport";
	***REMOVED***

	/**
	 * @param {ModuleGraph***REMOVED*** moduleGraph module graph
	 * @returns {null | false | GetConditionFn***REMOVED*** function to determine if the connection is active
	 */
	getCondition(moduleGraph) {
		return false;
	***REMOVED***
***REMOVED***

module.exports = LoaderImportDependency;
