// webpack/lib/dependencies/LocalModule.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const makeSerializable = require("../util/makeSerializable");

/** @typedef {import("../serialization/ObjectMiddleware").ObjectDeserializerContext***REMOVED*** ObjectDeserializerContext */
/** @typedef {import("../serialization/ObjectMiddleware").ObjectSerializerContext***REMOVED*** ObjectSerializerContext */

class LocalModule {
	/**
	 * @param {string***REMOVED*** name name
	 * @param {number***REMOVED*** idx index
	 */
	constructor(name, idx) {
		this.name = name;
		this.idx = idx;
		this.used = false;
	***REMOVED***

	flagUsed() {
		this.used = true;
	***REMOVED***

	/**
	 * @returns {string***REMOVED*** variable name
	 */
	variableName() {
		return `__WEBPACK_LOCAL_MODULE_${this.idx***REMOVED***__`;
	***REMOVED***

	/**
	 * @param {ObjectSerializerContext***REMOVED*** context context
	 */
	serialize(context) {
		const { write ***REMOVED*** = context;

		write(this.name);
		write(this.idx);
		write(this.used);
	***REMOVED***

	/**
	 * @param {ObjectDeserializerContext***REMOVED*** context context
	 */
	deserialize(context) {
		const { read ***REMOVED*** = context;

		this.name = read();
		this.idx = read();
		this.used = read();
	***REMOVED***
***REMOVED***

makeSerializable(LocalModule, "webpack/lib/dependencies/LocalModule");

module.exports = LocalModule;
