// es-abstract/operations/2016.js
'use strict';

module.exports = {
	IsPropertyDescriptor: 'https://262.ecma-international.org/6.0/#sec-property-descriptor-specification-type', // not actually an abstract op

	abs: {
		url: 'https://262.ecma-international.org/7.0/#sec-algorithm-conventions'
	***REMOVED***,
	'Abstract Equality Comparison': {
		url: 'https://262.ecma-international.org/7.0/#sec-abstract-equality-comparison'
	***REMOVED***,
	'Abstract Relational Comparison': {
		url: 'https://262.ecma-international.org/7.0/#sec-abstract-relational-comparison'
	***REMOVED***,
	AddRestrictedFunctionProperties: {
		url: 'https://262.ecma-international.org/7.0/#sec-addrestrictedfunctionproperties'
	***REMOVED***,
	AdvanceStringIndex: {
		url: 'https://262.ecma-international.org/7.0/#sec-advancestringindex'
	***REMOVED***,
	AllocateArrayBuffer: {
		url: 'https://262.ecma-international.org/7.0/#sec-allocatearraybuffer'
	***REMOVED***,
	AllocateTypedArray: {
		url: 'https://262.ecma-international.org/7.0/#sec-allocatetypedarray'
	***REMOVED***,
	AllocateTypedArrayBuffer: {
		url: 'https://262.ecma-international.org/7.0/#sec-allocatetypedarraybuffer'
	***REMOVED***,
	ArrayCreate: {
		url: 'https://262.ecma-international.org/7.0/#sec-arraycreate'
	***REMOVED***,
	ArraySetLength: {
		url: 'https://262.ecma-international.org/7.0/#sec-arraysetlength'
	***REMOVED***,
	ArraySpeciesCreate: {
		url: 'https://262.ecma-international.org/7.0/#sec-arrayspeciescreate'
	***REMOVED***,
	BlockDeclarationInstantiation: {
		url: 'https://262.ecma-international.org/7.0/#sec-blockdeclarationinstantiation'
	***REMOVED***,
	BoundFunctionCreate: {
		url: 'https://262.ecma-international.org/7.0/#sec-boundfunctioncreate'
	***REMOVED***,
	Call: {
		url: 'https://262.ecma-international.org/7.0/#sec-call'
	***REMOVED***,
	Canonicalize: {
		url: 'https://262.ecma-international.org/7.0/#sec-runtime-semantics-canonicalize-ch'
	***REMOVED***,
	CanonicalNumericIndexString: {
		url: 'https://262.ecma-international.org/7.0/#sec-canonicalnumericindexstring'
	***REMOVED***,
	CharacterRange: {
		url: 'https://262.ecma-international.org/7.0/#sec-runtime-semantics-characterrange-abstract-operation'
	***REMOVED***,
	CharacterRangeOrUnion: {
		url: 'https://262.ecma-international.org/7.0/#sec-runtime-semantics-characterrangeorunion-abstract-operation'
	***REMOVED***,
	CharacterSetMatcher: {
		url: 'https://262.ecma-international.org/7.0/#sec-runtime-semantics-charactersetmatcher-abstract-operation'
	***REMOVED***,
	CloneArrayBuffer: {
		url: 'https://262.ecma-international.org/7.0/#sec-clonearraybuffer'
	***REMOVED***,
	CompletePropertyDescriptor: {
		url: 'https://262.ecma-international.org/7.0/#sec-completepropertydescriptor'
	***REMOVED***,
	Completion: {
		url: 'https://262.ecma-international.org/7.0/#sec-completion-record-specification-type'
	***REMOVED***,
	CompletionRecord: {
		url: 'https://262.ecma-international.org/7.0/#sec-completion-record-specification-type'
	***REMOVED***,
	Construct: {
		url: 'https://262.ecma-international.org/7.0/#sec-construct'
	***REMOVED***,
	CopyDataBlockBytes: {
		url: 'https://262.ecma-international.org/7.0/#sec-copydatablockbytes'
	***REMOVED***,
	CreateArrayFromList: {
		url: 'https://262.ecma-international.org/7.0/#sec-createarrayfromlist'
	***REMOVED***,
	CreateArrayIterator: {
		url: 'https://262.ecma-international.org/7.0/#sec-createarrayiterator'
	***REMOVED***,
	CreateBuiltinFunction: {
		url: 'https://262.ecma-international.org/7.0/#sec-createbuiltinfunction'
	***REMOVED***,
	CreateByteDataBlock: {
		url: 'https://262.ecma-international.org/7.0/#sec-createbytedatablock'
	***REMOVED***,
	CreateDataProperty: {
		url: 'https://262.ecma-international.org/7.0/#sec-createdataproperty'
	***REMOVED***,
	CreateDataPropertyOrThrow: {
		url: 'https://262.ecma-international.org/7.0/#sec-createdatapropertyorthrow'
	***REMOVED***,
	CreateDynamicFunction: {
		url: 'https://262.ecma-international.org/7.0/#sec-createdynamicfunction'
	***REMOVED***,
	CreateHTML: {
		url: 'https://262.ecma-international.org/7.0/#sec-createhtml'
	***REMOVED***,
	CreateIntrinsics: {
		url: 'https://262.ecma-international.org/7.0/#sec-createintrinsics'
	***REMOVED***,
	CreateIterResultObject: {
		url: 'https://262.ecma-international.org/7.0/#sec-createiterresultobject'
	***REMOVED***,
	CreateListFromArrayLike: {
		url: 'https://262.ecma-international.org/7.0/#sec-createlistfromarraylike'
	***REMOVED***,
	CreateListIterator: {
		url: 'https://262.ecma-international.org/7.0/#sec-createlistiterator'
	***REMOVED***,
	CreateMapIterator: {
		url: 'https://262.ecma-international.org/7.0/#sec-createmapiterator'
	***REMOVED***,
	CreateMappedArgumentsObject: {
		url: 'https://262.ecma-international.org/7.0/#sec-createmappedargumentsobject'
	***REMOVED***,
	CreateMethodProperty: {
		url: 'https://262.ecma-international.org/7.0/#sec-createmethodproperty'
	***REMOVED***,
	CreatePerIterationEnvironment: {
		url: 'https://262.ecma-international.org/7.0/#sec-createperiterationenvironment'
	***REMOVED***,
	CreateRealm: {
		url: 'https://262.ecma-international.org/7.0/#sec-createrealm'
	***REMOVED***,
	CreateResolvingFunctions: {
		url: 'https://262.ecma-international.org/7.0/#sec-createresolvingfunctions'
	***REMOVED***,
	CreateSetIterator: {
		url: 'https://262.ecma-international.org/7.0/#sec-createsetiterator'
	***REMOVED***,
	CreateStringIterator: {
		url: 'https://262.ecma-international.org/7.0/#sec-createstringiterator'
	***REMOVED***,
	CreateUnmappedArgumentsObject: {
		url: 'https://262.ecma-international.org/7.0/#sec-createunmappedargumentsobject'
	***REMOVED***,
	DateFromTime: {
		url: 'https://262.ecma-international.org/7.0/#sec-date-number'
	***REMOVED***,
	Day: {
		url: 'https://262.ecma-international.org/7.0/#sec-day-number-and-time-within-day'
	***REMOVED***,
	DayFromYear: {
		url: 'https://262.ecma-international.org/7.0/#sec-year-number'
	***REMOVED***,
	DaylightSavingTA: {
		url: 'https://262.ecma-international.org/7.0/#sec-daylight-saving-time-adjustment'
	***REMOVED***,
	DaysInYear: {
		url: 'https://262.ecma-international.org/7.0/#sec-year-number'
	***REMOVED***,
	DayWithinYear: {
		url: 'https://262.ecma-international.org/7.0/#sec-month-number'
	***REMOVED***,
	Decode: {
		url: 'https://262.ecma-international.org/7.0/#sec-decode'
	***REMOVED***,
	DefinePropertyOrThrow: {
		url: 'https://262.ecma-international.org/7.0/#sec-definepropertyorthrow'
	***REMOVED***,
	DeletePropertyOrThrow: {
		url: 'https://262.ecma-international.org/7.0/#sec-deletepropertyorthrow'
	***REMOVED***,
	DetachArrayBuffer: {
		url: 'https://262.ecma-international.org/7.0/#sec-detacharraybuffer'
	***REMOVED***,
	Encode: {
		url: 'https://262.ecma-international.org/7.0/#sec-encode'
	***REMOVED***,
	EnqueueJob: {
		url: 'https://262.ecma-international.org/7.0/#sec-enqueuejob'
	***REMOVED***,
	EnumerableOwnNames: {
		url: 'https://262.ecma-international.org/7.0/#sec-enumerableownnames'
	***REMOVED***,
	EnumerateObjectProperties: {
		url: 'https://262.ecma-international.org/7.0/#sec-enumerate-object-properties'
	***REMOVED***,
	EscapeRegExpPattern: {
		url: 'https://262.ecma-international.org/7.0/#sec-escaperegexppattern'
	***REMOVED***,
	EvalDeclarationInstantiation: {
		url: 'https://262.ecma-international.org/7.0/#sec-evaldeclarationinstantiation'
	***REMOVED***,
	EvaluateCall: {
		url: 'https://262.ecma-international.org/7.0/#sec-evaluatecall'
	***REMOVED***,
	EvaluateDirectCall: {
		url: 'https://262.ecma-international.org/7.0/#sec-evaluatedirectcall'
	***REMOVED***,
	EvaluateNew: {
		url: 'https://262.ecma-international.org/7.0/#sec-evaluatenew'
	***REMOVED***,
	floor: {
		url: 'https://262.ecma-international.org/7.0/#sec-algorithm-conventions'
	***REMOVED***,
	ForBodyEvaluation: {
		url: 'https://262.ecma-international.org/7.0/#sec-forbodyevaluation'
	***REMOVED***,
	'ForIn/OfBodyEvaluation': {
		url: 'https://262.ecma-international.org/7.0/#sec-runtime-semantics-forin-div-ofbodyevaluation-lhs-stmt-iterator-lhskind-labelset'
	***REMOVED***,
	'ForIn/OfHeadEvaluation': {
		url: 'https://262.ecma-international.org/7.0/#sec-runtime-semantics-forin-div-ofheadevaluation-tdznames-expr-iterationkind'
	***REMOVED***,
	FromPropertyDescriptor: {
		url: 'https://262.ecma-international.org/7.0/#sec-frompropertydescriptor'
	***REMOVED***,
	FulfillPromise: {
		url: 'https://262.ecma-international.org/7.0/#sec-fulfillpromise'
	***REMOVED***,
	FunctionAllocate: {
		url: 'https://262.ecma-international.org/7.0/#sec-functionallocate'
	***REMOVED***,
	FunctionCreate: {
		url: 'https://262.ecma-international.org/7.0/#sec-functioncreate'
	***REMOVED***,
	FunctionDeclarationInstantiation: {
		url: 'https://262.ecma-international.org/7.0/#sec-functiondeclarationinstantiation'
	***REMOVED***,
	FunctionInitialize: {
		url: 'https://262.ecma-international.org/7.0/#sec-functioninitialize'
	***REMOVED***,
	GeneratorFunctionCreate: {
		url: 'https://262.ecma-international.org/7.0/#sec-generatorfunctioncreate'
	***REMOVED***,
	GeneratorResume: {
		url: 'https://262.ecma-international.org/7.0/#sec-generatorresume'
	***REMOVED***,
	GeneratorResumeAbrupt: {
		url: 'https://262.ecma-international.org/7.0/#sec-generatorresumeabrupt'
	***REMOVED***,
	GeneratorStart: {
		url: 'https://262.ecma-international.org/7.0/#sec-generatorstart'
	***REMOVED***,
	GeneratorValidate: {
		url: 'https://262.ecma-international.org/7.0/#sec-generatorvalidate'
	***REMOVED***,
	GeneratorYield: {
		url: 'https://262.ecma-international.org/7.0/#sec-generatoryield'
	***REMOVED***,
	Get: {
		url: 'https://262.ecma-international.org/7.0/#sec-get-o-p'
	***REMOVED***,
	GetActiveScriptOrModule: {
		url: 'https://262.ecma-international.org/7.0/#sec-getactivescriptormodule'
	***REMOVED***,
	GetFunctionRealm: {
		url: 'https://262.ecma-international.org/7.0/#sec-getfunctionrealm'
	***REMOVED***,
	GetGlobalObject: {
		url: 'https://262.ecma-international.org/7.0/#sec-getglobalobject'
	***REMOVED***,
	GetIdentifierReference: {
		url: 'https://262.ecma-international.org/7.0/#sec-getidentifierreference'
	***REMOVED***,
	GetIterator: {
		url: 'https://262.ecma-international.org/7.0/#sec-getiterator'
	***REMOVED***,
	GetMethod: {
		url: 'https://262.ecma-international.org/7.0/#sec-getmethod'
	***REMOVED***,
	GetModuleNamespace: {
		url: 'https://262.ecma-international.org/7.0/#sec-getmodulenamespace'
	***REMOVED***,
	GetNewTarget: {
		url: 'https://262.ecma-international.org/7.0/#sec-getnewtarget'
	***REMOVED***,
	GetOwnPropertyKeys: {
		url: 'https://262.ecma-international.org/7.0/#sec-getownpropertykeys'
	***REMOVED***,
	GetPrototypeFromConstructor: {
		url: 'https://262.ecma-international.org/7.0/#sec-getprototypefromconstructor'
	***REMOVED***,
	GetSubstitution: {
		url: 'https://262.ecma-international.org/7.0/#sec-getsubstitution'
	***REMOVED***,
	GetSuperConstructor: {
		url: 'https://262.ecma-international.org/7.0/#sec-getsuperconstructor'
	***REMOVED***,
	GetTemplateObject: {
		url: 'https://262.ecma-international.org/7.0/#sec-gettemplateobject'
	***REMOVED***,
	GetThisEnvironment: {
		url: 'https://262.ecma-international.org/7.0/#sec-getthisenvironment'
	***REMOVED***,
	GetThisValue: {
		url: 'https://262.ecma-international.org/7.0/#sec-getthisvalue'
	***REMOVED***,
	GetV: {
		url: 'https://262.ecma-international.org/7.0/#sec-getv'
	***REMOVED***,
	GetValue: {
		url: 'https://262.ecma-international.org/7.0/#sec-getvalue'
	***REMOVED***,
	GetValueFromBuffer: {
		url: 'https://262.ecma-international.org/7.0/#sec-getvaluefrombuffer'
	***REMOVED***,
	GetViewValue: {
		url: 'https://262.ecma-international.org/7.0/#sec-getviewvalue'
	***REMOVED***,
	GlobalDeclarationInstantiation: {
		url: 'https://262.ecma-international.org/7.0/#sec-globaldeclarationinstantiation'
	***REMOVED***,
	HasOwnProperty: {
		url: 'https://262.ecma-international.org/7.0/#sec-hasownproperty'
	***REMOVED***,
	HasProperty: {
		url: 'https://262.ecma-international.org/7.0/#sec-hasproperty'
	***REMOVED***,
	HourFromTime: {
		url: 'https://262.ecma-international.org/7.0/#sec-hours-minutes-second-and-milliseconds'
	***REMOVED***,
	IfAbruptRejectPromise: {
		url: 'https://262.ecma-international.org/7.0/#sec-ifabruptrejectpromise'
	***REMOVED***,
	ImportedLocalNames: {
		url: 'https://262.ecma-international.org/7.0/#sec-importedlocalnames'
	***REMOVED***,
	InitializeBoundName: {
		url: 'https://262.ecma-international.org/7.0/#sec-initializeboundname'
	***REMOVED***,
	InitializeHostDefinedRealm: {
		url: 'https://262.ecma-international.org/7.0/#sec-initializehostdefinedrealm'
	***REMOVED***,
	InitializeReferencedBinding: {
		url: 'https://262.ecma-international.org/7.0/#sec-initializereferencedbinding'
	***REMOVED***,
	InLeapYear: {
		url: 'https://262.ecma-international.org/7.0/#sec-year-number'
	***REMOVED***,
	InstanceofOperator: {
		url: 'https://262.ecma-international.org/7.0/#sec-instanceofoperator'
	***REMOVED***,
	IntegerIndexedElementGet: {
		url: 'https://262.ecma-international.org/7.0/#sec-integerindexedelementget'
	***REMOVED***,
	IntegerIndexedElementSet: {
		url: 'https://262.ecma-international.org/7.0/#sec-integerindexedelementset'
	***REMOVED***,
	IntegerIndexedObjectCreate: {
		url: 'https://262.ecma-international.org/7.0/#sec-integerindexedobjectcreate'
	***REMOVED***,
	InternalizeJSONProperty: {
		url: 'https://262.ecma-international.org/7.0/#sec-internalizejsonproperty'
	***REMOVED***,
	Invoke: {
		url: 'https://262.ecma-international.org/7.0/#sec-invoke'
	***REMOVED***,
	IsAccessorDescriptor: {
		url: 'https://262.ecma-international.org/7.0/#sec-isaccessordescriptor'
	***REMOVED***,
	IsAnonymousFunctionDefinition: {
		url: 'https://262.ecma-international.org/7.0/#sec-isanonymousfunctiondefinition'
	***REMOVED***,
	IsArray: {
		url: 'https://262.ecma-international.org/7.0/#sec-isarray'
	***REMOVED***,
	IsCallable: {
		url: 'https://262.ecma-international.org/7.0/#sec-iscallable'
	***REMOVED***,
	IsCompatiblePropertyDescriptor: {
		url: 'https://262.ecma-international.org/7.0/#sec-iscompatiblepropertydescriptor'
	***REMOVED***,
	IsConcatSpreadable: {
		url: 'https://262.ecma-international.org/7.0/#sec-isconcatspreadable'
	***REMOVED***,
	IsConstructor: {
		url: 'https://262.ecma-international.org/7.0/#sec-isconstructor'
	***REMOVED***,
	IsDataDescriptor: {
		url: 'https://262.ecma-international.org/7.0/#sec-isdatadescriptor'
	***REMOVED***,
	IsDetachedBuffer: {
		url: 'https://262.ecma-international.org/7.0/#sec-isdetachedbuffer'
	***REMOVED***,
	IsExtensible: {
		url: 'https://262.ecma-international.org/7.0/#sec-isextensible-o'
	***REMOVED***,
	IsGenericDescriptor: {
		url: 'https://262.ecma-international.org/7.0/#sec-isgenericdescriptor'
	***REMOVED***,
	IsInTailPosition: {
		url: 'https://262.ecma-international.org/7.0/#sec-isintailposition'
	***REMOVED***,
	IsInteger: {
		url: 'https://262.ecma-international.org/7.0/#sec-isinteger'
	***REMOVED***,
	IsLabelledFunction: {
		url: 'https://262.ecma-international.org/7.0/#sec-islabelledfunction'
	***REMOVED***,
	IsPromise: {
		url: 'https://262.ecma-international.org/7.0/#sec-ispromise'
	***REMOVED***,
	IsPropertyKey: {
		url: 'https://262.ecma-international.org/7.0/#sec-ispropertykey'
	***REMOVED***,
	IsRegExp: {
		url: 'https://262.ecma-international.org/7.0/#sec-isregexp'
	***REMOVED***,
	IsWordChar: {
		url: 'https://262.ecma-international.org/7.0/#sec-runtime-semantics-iswordchar-abstract-operation'
	***REMOVED***,
	IterableToArrayLike: {
		url: 'https://262.ecma-international.org/7.0/#sec-iterabletoarraylike'
	***REMOVED***,
	IteratorClose: {
		url: 'https://262.ecma-international.org/7.0/#sec-iteratorclose'
	***REMOVED***,
	IteratorComplete: {
		url: 'https://262.ecma-international.org/7.0/#sec-iteratorcomplete'
	***REMOVED***,
	IteratorNext: {
		url: 'https://262.ecma-international.org/7.0/#sec-iteratornext'
	***REMOVED***,
	IteratorStep: {
		url: 'https://262.ecma-international.org/7.0/#sec-iteratorstep'
	***REMOVED***,
	IteratorValue: {
		url: 'https://262.ecma-international.org/7.0/#sec-iteratorvalue'
	***REMOVED***,
	LocalTime: {
		url: 'https://262.ecma-international.org/7.0/#sec-localtime'
	***REMOVED***,
	LoopContinues: {
		url: 'https://262.ecma-international.org/7.0/#sec-loopcontinues'
	***REMOVED***,
	MakeArgGetter: {
		url: 'https://262.ecma-international.org/7.0/#sec-makearggetter'
	***REMOVED***,
	MakeArgSetter: {
		url: 'https://262.ecma-international.org/7.0/#sec-makeargsetter'
	***REMOVED***,
	MakeClassConstructor: {
		url: 'https://262.ecma-international.org/7.0/#sec-makeclassconstructor'
	***REMOVED***,
	MakeConstructor: {
		url: 'https://262.ecma-international.org/7.0/#sec-makeconstructor'
	***REMOVED***,
	MakeDate: {
		url: 'https://262.ecma-international.org/7.0/#sec-makedate'
	***REMOVED***,
	MakeDay: {
		url: 'https://262.ecma-international.org/7.0/#sec-makeday'
	***REMOVED***,
	MakeMethod: {
		url: 'https://262.ecma-international.org/7.0/#sec-makemethod'
	***REMOVED***,
	MakeSuperPropertyReference: {
		url: 'https://262.ecma-international.org/7.0/#sec-makesuperpropertyreference'
	***REMOVED***,
	MakeTime: {
		url: 'https://262.ecma-international.org/7.0/#sec-maketime'
	***REMOVED***,
	max: {
		url: 'https://262.ecma-international.org/7.0/#sec-algorithm-conventions'
	***REMOVED***,
	min: {
		url: 'https://262.ecma-international.org/7.0/#sec-algorithm-conventions'
	***REMOVED***,
	MinFromTime: {
		url: 'https://262.ecma-international.org/7.0/#sec-hours-minutes-second-and-milliseconds'
	***REMOVED***,
	ModuleNamespaceCreate: {
		url: 'https://262.ecma-international.org/7.0/#sec-modulenamespacecreate'
	***REMOVED***,
	modulo: {
		url: 'https://262.ecma-international.org/7.0/#sec-algorithm-conventions'
	***REMOVED***,
	MonthFromTime: {
		url: 'https://262.ecma-international.org/7.0/#sec-month-number'
	***REMOVED***,
	msFromTime: {
		url: 'https://262.ecma-international.org/7.0/#sec-hours-minutes-second-and-milliseconds'
	***REMOVED***,
	NewDeclarativeEnvironment: {
		url: 'https://262.ecma-international.org/7.0/#sec-newdeclarativeenvironment'
	***REMOVED***,
	NewFunctionEnvironment: {
		url: 'https://262.ecma-international.org/7.0/#sec-newfunctionenvironment'
	***REMOVED***,
	NewGlobalEnvironment: {
		url: 'https://262.ecma-international.org/7.0/#sec-newglobalenvironment'
	***REMOVED***,
	NewModuleEnvironment: {
		url: 'https://262.ecma-international.org/7.0/#sec-newmoduleenvironment'
	***REMOVED***,
	NewObjectEnvironment: {
		url: 'https://262.ecma-international.org/7.0/#sec-newobjectenvironment'
	***REMOVED***,
	NewPromiseCapability: {
		url: 'https://262.ecma-international.org/7.0/#sec-newpromisecapability'
	***REMOVED***,
	NextJob: {
		url: 'https://262.ecma-international.org/7.0/#sec-nextjob-result'
	***REMOVED***,
	NormalCompletion: {
		url: 'https://262.ecma-international.org/7.0/#sec-normalcompletion'
	***REMOVED***,
	ObjectCreate: {
		url: 'https://262.ecma-international.org/7.0/#sec-objectcreate'
	***REMOVED***,
	ObjectDefineProperties: {
		url: 'https://262.ecma-international.org/7.0/#sec-objectdefineproperties'
	***REMOVED***,
	OrdinaryCallBindThis: {
		url: 'https://262.ecma-international.org/7.0/#sec-ordinarycallbindthis'
	***REMOVED***,
	OrdinaryCallEvaluateBody: {
		url: 'https://262.ecma-international.org/7.0/#sec-ordinarycallevaluatebody'
	***REMOVED***,
	OrdinaryCreateFromConstructor: {
		url: 'https://262.ecma-international.org/7.0/#sec-ordinarycreatefromconstructor'
	***REMOVED***,
	OrdinaryDefineOwnProperty: {
		url: 'https://262.ecma-international.org/7.0/#sec-ordinarydefineownproperty'
	***REMOVED***,
	OrdinaryDelete: {
		url: 'https://262.ecma-international.org/7.0/#sec-ordinarydelete'
	***REMOVED***,
	OrdinaryGet: {
		url: 'https://262.ecma-international.org/7.0/#sec-ordinaryget'
	***REMOVED***,
	OrdinaryGetOwnProperty: {
		url: 'https://262.ecma-international.org/7.0/#sec-ordinarygetownproperty'
	***REMOVED***,
	OrdinaryGetPrototypeOf: {
		url: 'https://262.ecma-international.org/7.0/#sec-ordinarygetprototypeof'
	***REMOVED***,
	OrdinaryHasInstance: {
		url: 'https://262.ecma-international.org/7.0/#sec-ordinaryhasinstance'
	***REMOVED***,
	OrdinaryHasProperty: {
		url: 'https://262.ecma-international.org/7.0/#sec-ordinaryhasproperty'
	***REMOVED***,
	OrdinaryIsExtensible: {
		url: 'https://262.ecma-international.org/7.0/#sec-ordinaryisextensible'
	***REMOVED***,
	OrdinaryOwnPropertyKeys: {
		url: 'https://262.ecma-international.org/7.0/#sec-ordinaryownpropertykeys'
	***REMOVED***,
	OrdinaryPreventExtensions: {
		url: 'https://262.ecma-international.org/7.0/#sec-ordinarypreventextensions'
	***REMOVED***,
	OrdinarySet: {
		url: 'https://262.ecma-international.org/7.0/#sec-ordinaryset'
	***REMOVED***,
	OrdinarySetPrototypeOf: {
		url: 'https://262.ecma-international.org/7.0/#sec-ordinarysetprototypeof'
	***REMOVED***,
	ParseModule: {
		url: 'https://262.ecma-international.org/7.0/#sec-parsemodule'
	***REMOVED***,
	ParseScript: {
		url: 'https://262.ecma-international.org/7.0/#sec-parse-script'
	***REMOVED***,
	PerformEval: {
		url: 'https://262.ecma-international.org/7.0/#sec-performeval'
	***REMOVED***,
	PerformPromiseAll: {
		url: 'https://262.ecma-international.org/7.0/#sec-performpromiseall'
	***REMOVED***,
	PerformPromiseRace: {
		url: 'https://262.ecma-international.org/7.0/#sec-performpromiserace'
	***REMOVED***,
	PerformPromiseThen: {
		url: 'https://262.ecma-international.org/7.0/#sec-performpromisethen'
	***REMOVED***,
	PrepareForOrdinaryCall: {
		url: 'https://262.ecma-international.org/7.0/#sec-prepareforordinarycall'
	***REMOVED***,
	PrepareForTailCall: {
		url: 'https://262.ecma-international.org/7.0/#sec-preparefortailcall'
	***REMOVED***,
	PromiseReactionJob: {
		url: 'https://262.ecma-international.org/7.0/#sec-promisereactionjob'
	***REMOVED***,
	PromiseResolveThenableJob: {
		url: 'https://262.ecma-international.org/7.0/#sec-promiseresolvethenablejob'
	***REMOVED***,
	ProxyCreate: {
		url: 'https://262.ecma-international.org/7.0/#sec-proxycreate'
	***REMOVED***,
	PutValue: {
		url: 'https://262.ecma-international.org/7.0/#sec-putvalue'
	***REMOVED***,
	QuoteJSONString: {
		url: 'https://262.ecma-international.org/7.0/#sec-quotejsonstring'
	***REMOVED***,
	RegExpAlloc: {
		url: 'https://262.ecma-international.org/7.0/#sec-regexpalloc'
	***REMOVED***,
	RegExpBuiltinExec: {
		url: 'https://262.ecma-international.org/7.0/#sec-regexpbuiltinexec'
	***REMOVED***,
	RegExpCreate: {
		url: 'https://262.ecma-international.org/7.0/#sec-regexpcreate'
	***REMOVED***,
	RegExpExec: {
		url: 'https://262.ecma-international.org/7.0/#sec-regexpexec'
	***REMOVED***,
	RegExpInitialize: {
		url: 'https://262.ecma-international.org/7.0/#sec-regexpinitialize'
	***REMOVED***,
	RejectPromise: {
		url: 'https://262.ecma-international.org/7.0/#sec-rejectpromise'
	***REMOVED***,
	RepeatMatcher: {
		url: 'https://262.ecma-international.org/7.0/#sec-runtime-semantics-repeatmatcher-abstract-operation'
	***REMOVED***,
	RequireObjectCoercible: {
		url: 'https://262.ecma-international.org/7.0/#sec-requireobjectcoercible'
	***REMOVED***,
	ResolveBinding: {
		url: 'https://262.ecma-international.org/7.0/#sec-resolvebinding'
	***REMOVED***,
	ResolveThisBinding: {
		url: 'https://262.ecma-international.org/7.0/#sec-resolvethisbinding'
	***REMOVED***,
	ReturnIfAbrupt: {
		url: 'https://262.ecma-international.org/7.0/#sec-returnifabrupt'
	***REMOVED***,
	SameValue: {
		url: 'https://262.ecma-international.org/7.0/#sec-samevalue'
	***REMOVED***,
	SameValueNonNumber: {
		url: 'https://262.ecma-international.org/7.0/#sec-samevaluenonnumber'
	***REMOVED***,
	SameValueZero: {
		url: 'https://262.ecma-international.org/7.0/#sec-samevaluezero'
	***REMOVED***,
	ScriptEvaluation: {
		url: 'https://262.ecma-international.org/7.0/#sec-runtime-semantics-scriptevaluation'
	***REMOVED***,
	ScriptEvaluationJob: {
		url: 'https://262.ecma-international.org/7.0/#sec-scriptevaluationjob'
	***REMOVED***,
	SecFromTime: {
		url: 'https://262.ecma-international.org/7.0/#sec-hours-minutes-second-and-milliseconds'
	***REMOVED***,
	SerializeJSONArray: {
		url: 'https://262.ecma-international.org/7.0/#sec-serializejsonarray'
	***REMOVED***,
	SerializeJSONObject: {
		url: 'https://262.ecma-international.org/7.0/#sec-serializejsonobject'
	***REMOVED***,
	SerializeJSONProperty: {
		url: 'https://262.ecma-international.org/7.0/#sec-serializejsonproperty'
	***REMOVED***,
	Set: {
		url: 'https://262.ecma-international.org/7.0/#sec-set-o-p-v-throw'
	***REMOVED***,
	SetDefaultGlobalBindings: {
		url: 'https://262.ecma-international.org/7.0/#sec-setdefaultglobalbindings'
	***REMOVED***,
	SetFunctionName: {
		url: 'https://262.ecma-international.org/7.0/#sec-setfunctionname'
	***REMOVED***,
	SetIntegrityLevel: {
		url: 'https://262.ecma-international.org/7.0/#sec-setintegritylevel'
	***REMOVED***,
	SetRealmGlobalObject: {
		url: 'https://262.ecma-international.org/7.0/#sec-setrealmglobalobject'
	***REMOVED***,
	SetValueInBuffer: {
		url: 'https://262.ecma-international.org/7.0/#sec-setvalueinbuffer'
	***REMOVED***,
	SetViewValue: {
		url: 'https://262.ecma-international.org/7.0/#sec-setviewvalue'
	***REMOVED***,
	SortCompare: {
		url: 'https://262.ecma-international.org/7.0/#sec-sortcompare'
	***REMOVED***,
	SpeciesConstructor: {
		url: 'https://262.ecma-international.org/7.0/#sec-speciesconstructor'
	***REMOVED***,
	SplitMatch: {
		url: 'https://262.ecma-international.org/7.0/#sec-splitmatch'
	***REMOVED***,
	'Strict Equality Comparison': {
		url: 'https://262.ecma-international.org/7.0/#sec-strict-equality-comparison'
	***REMOVED***,
	StringCreate: {
		url: 'https://262.ecma-international.org/7.0/#sec-stringcreate'
	***REMOVED***,
	SymbolDescriptiveString: {
		url: 'https://262.ecma-international.org/7.0/#sec-symboldescriptivestring'
	***REMOVED***,
	TestIntegrityLevel: {
		url: 'https://262.ecma-international.org/7.0/#sec-testintegritylevel'
	***REMOVED***,
	thisBooleanValue: {
		url: 'https://262.ecma-international.org/7.0/#sec-thisbooleanvalue'
	***REMOVED***,
	thisNumberValue: {
		url: 'https://262.ecma-international.org/7.0/#sec-properties-of-the-number-prototype-object'
	***REMOVED***,
	thisStringValue: {
		url: 'https://262.ecma-international.org/7.0/#sec-properties-of-the-string-prototype-object'
	***REMOVED***,
	thisTimeValue: {
		url: 'https://262.ecma-international.org/7.0/#sec-properties-of-the-date-prototype-object'
	***REMOVED***,
	TimeClip: {
		url: 'https://262.ecma-international.org/7.0/#sec-timeclip'
	***REMOVED***,
	TimeFromYear: {
		url: 'https://262.ecma-international.org/7.0/#sec-year-number'
	***REMOVED***,
	TimeWithinDay: {
		url: 'https://262.ecma-international.org/7.0/#sec-day-number-and-time-within-day'
	***REMOVED***,
	ToBoolean: {
		url: 'https://262.ecma-international.org/7.0/#sec-toboolean'
	***REMOVED***,
	ToDateString: {
		url: 'https://262.ecma-international.org/7.0/#sec-todatestring'
	***REMOVED***,
	ToInt16: {
		url: 'https://262.ecma-international.org/7.0/#sec-toint16'
	***REMOVED***,
	ToInt32: {
		url: 'https://262.ecma-international.org/7.0/#sec-toint32'
	***REMOVED***,
	ToInt8: {
		url: 'https://262.ecma-international.org/7.0/#sec-toint8'
	***REMOVED***,
	ToInteger: {
		url: 'https://262.ecma-international.org/7.0/#sec-tointeger'
	***REMOVED***,
	ToLength: {
		url: 'https://262.ecma-international.org/7.0/#sec-tolength'
	***REMOVED***,
	ToNumber: {
		url: 'https://262.ecma-international.org/7.0/#sec-tonumber'
	***REMOVED***,
	ToObject: {
		url: 'https://262.ecma-international.org/7.0/#sec-toobject'
	***REMOVED***,
	TopLevelModuleEvaluationJob: {
		url: 'https://262.ecma-international.org/7.0/#sec-toplevelmoduleevaluationjob'
	***REMOVED***,
	ToPrimitive: {
		url: 'https://262.ecma-international.org/7.0/#sec-toprimitive'
	***REMOVED***,
	ToPropertyDescriptor: {
		url: 'https://262.ecma-international.org/7.0/#sec-topropertydescriptor'
	***REMOVED***,
	ToPropertyKey: {
		url: 'https://262.ecma-international.org/7.0/#sec-topropertykey'
	***REMOVED***,
	ToString: {
		url: 'https://262.ecma-international.org/7.0/#sec-tostring'
	***REMOVED***,
	'ToString Applied to the Number Type': {
		url: 'https://262.ecma-international.org/7.0/#sec-tostring-applied-to-the-number-type'
	***REMOVED***,
	ToUint16: {
		url: 'https://262.ecma-international.org/7.0/#sec-touint16'
	***REMOVED***,
	ToUint32: {
		url: 'https://262.ecma-international.org/7.0/#sec-touint32'
	***REMOVED***,
	ToUint8: {
		url: 'https://262.ecma-international.org/7.0/#sec-touint8'
	***REMOVED***,
	ToUint8Clamp: {
		url: 'https://262.ecma-international.org/7.0/#sec-touint8clamp'
	***REMOVED***,
	TriggerPromiseReactions: {
		url: 'https://262.ecma-international.org/7.0/#sec-triggerpromisereactions'
	***REMOVED***,
	Type: {
		url: 'https://262.ecma-international.org/7.0/#sec-ecmascript-data-types-and-values'
	***REMOVED***,
	TypedArrayCreate: {
		url: 'https://262.ecma-international.org/7.0/#typedarray-create'
	***REMOVED***,
	TypedArraySpeciesCreate: {
		url: 'https://262.ecma-international.org/7.0/#typedarray-species-create'
	***REMOVED***,
	UpdateEmpty: {
		url: 'https://262.ecma-international.org/7.0/#sec-updateempty'
	***REMOVED***,
	UTC: {
		url: 'https://262.ecma-international.org/7.0/#sec-utc-t'
	***REMOVED***,
	UTF16Decode: {
		url: 'https://262.ecma-international.org/7.0/#sec-utf16decode'
	***REMOVED***,
	UTF16Encoding: {
		url: 'https://262.ecma-international.org/7.0/#sec-utf16encoding'
	***REMOVED***,
	ValidateAndApplyPropertyDescriptor: {
		url: 'https://262.ecma-international.org/7.0/#sec-validateandapplypropertydescriptor'
	***REMOVED***,
	ValidateTypedArray: {
		url: 'https://262.ecma-international.org/7.0/#sec-validatetypedarray'
	***REMOVED***,
	WeekDay: {
		url: 'https://262.ecma-international.org/7.0/#sec-week-day'
	***REMOVED***,
	YearFromTime: {
		url: 'https://262.ecma-international.org/7.0/#sec-year-number'
	***REMOVED***
***REMOVED***;
