// es-abstract/operations/2021.js
'use strict';

module.exports = {
	abs: {
		url: 'https://262.ecma-international.org/12.0/#eqn-abs'
	***REMOVED***,
	'Abstract Equality Comparison': {
		url: 'https://262.ecma-international.org/12.0/#sec-abstract-equality-comparison'
	***REMOVED***,
	'Abstract Relational Comparison': {
		url: 'https://262.ecma-international.org/12.0/#sec-abstract-relational-comparison'
	***REMOVED***,
	AddEntriesFromIterable: {
		url: 'https://262.ecma-international.org/12.0/#sec-add-entries-from-iterable'
	***REMOVED***,
	AddRestrictedFunctionProperties: {
		url: 'https://262.ecma-international.org/12.0/#sec-addrestrictedfunctionproperties'
	***REMOVED***,
	AddToKeptObjects: {
		url: 'https://262.ecma-international.org/12.0/#sec-addtokeptobjects'
	***REMOVED***,
	AddWaiter: {
		url: 'https://262.ecma-international.org/12.0/#sec-addwaiter'
	***REMOVED***,
	AdvanceStringIndex: {
		url: 'https://262.ecma-international.org/12.0/#sec-advancestringindex'
	***REMOVED***,
	'agent-order': {
		url: 'https://262.ecma-international.org/12.0/#sec-agent-order'
	***REMOVED***,
	AgentCanSuspend: {
		url: 'https://262.ecma-international.org/12.0/#sec-agentcansuspend'
	***REMOVED***,
	AgentSignifier: {
		url: 'https://262.ecma-international.org/12.0/#sec-agentsignifier'
	***REMOVED***,
	AllocateArrayBuffer: {
		url: 'https://262.ecma-international.org/12.0/#sec-allocatearraybuffer'
	***REMOVED***,
	AllocateSharedArrayBuffer: {
		url: 'https://262.ecma-international.org/12.0/#sec-allocatesharedarraybuffer'
	***REMOVED***,
	AllocateTypedArray: {
		url: 'https://262.ecma-international.org/12.0/#sec-allocatetypedarray'
	***REMOVED***,
	AllocateTypedArrayBuffer: {
		url: 'https://262.ecma-international.org/12.0/#sec-allocatetypedarraybuffer'
	***REMOVED***,
	ApplyStringOrNumericBinaryOperator: {
		url: 'https://262.ecma-international.org/12.0/#sec-applystringornumericbinaryoperator'
	***REMOVED***,
	ArrayCreate: {
		url: 'https://262.ecma-international.org/12.0/#sec-arraycreate'
	***REMOVED***,
	ArraySetLength: {
		url: 'https://262.ecma-international.org/12.0/#sec-arraysetlength'
	***REMOVED***,
	ArraySpeciesCreate: {
		url: 'https://262.ecma-international.org/12.0/#sec-arrayspeciescreate'
	***REMOVED***,
	AsyncFromSyncIteratorContinuation: {
		url: 'https://262.ecma-international.org/12.0/#sec-asyncfromsynciteratorcontinuation'
	***REMOVED***,
	AsyncFunctionStart: {
		url: 'https://262.ecma-international.org/12.0/#sec-async-functions-abstract-operations-async-function-start'
	***REMOVED***,
	AsyncGeneratorEnqueue: {
		url: 'https://262.ecma-international.org/12.0/#sec-asyncgeneratorenqueue'
	***REMOVED***,
	AsyncGeneratorReject: {
		url: 'https://262.ecma-international.org/12.0/#sec-asyncgeneratorreject'
	***REMOVED***,
	AsyncGeneratorResolve: {
		url: 'https://262.ecma-international.org/12.0/#sec-asyncgeneratorresolve'
	***REMOVED***,
	AsyncGeneratorResumeNext: {
		url: 'https://262.ecma-international.org/12.0/#sec-asyncgeneratorresumenext'
	***REMOVED***,
	AsyncGeneratorStart: {
		url: 'https://262.ecma-international.org/12.0/#sec-asyncgeneratorstart'
	***REMOVED***,
	AsyncGeneratorValidate: {
		url: 'https://262.ecma-international.org/12.0/#sec-asyncgeneratorvalidate'
	***REMOVED***,
	AsyncGeneratorYield: {
		url: 'https://262.ecma-international.org/12.0/#sec-asyncgeneratoryield'
	***REMOVED***,
	AsyncIteratorClose: {
		url: 'https://262.ecma-international.org/12.0/#sec-asynciteratorclose'
	***REMOVED***,
	AtomicReadModifyWrite: {
		url: 'https://262.ecma-international.org/12.0/#sec-atomicreadmodifywrite'
	***REMOVED***,
	Await: {
		url: 'https://262.ecma-international.org/12.0/#await'
	***REMOVED***,
	BackreferenceMatcher: {
		url: 'https://262.ecma-international.org/12.0/#sec-backreference-matcher'
	***REMOVED***,
	'BigInt::add': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-add'
	***REMOVED***,
	'BigInt::bitwiseAND': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-bitwiseAND'
	***REMOVED***,
	'BigInt::bitwiseNOT': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-bitwiseNOT'
	***REMOVED***,
	'BigInt::bitwiseOR': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-bitwiseOR'
	***REMOVED***,
	'BigInt::bitwiseXOR': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-bitwiseXOR'
	***REMOVED***,
	'BigInt::divide': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-divide'
	***REMOVED***,
	'BigInt::equal': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-equal'
	***REMOVED***,
	'BigInt::exponentiate': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-exponentiate'
	***REMOVED***,
	'BigInt::leftShift': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-leftShift'
	***REMOVED***,
	'BigInt::lessThan': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-lessThan'
	***REMOVED***,
	'BigInt::multiply': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-multiply'
	***REMOVED***,
	'BigInt::remainder': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-remainder'
	***REMOVED***,
	'BigInt::sameValue': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-sameValue'
	***REMOVED***,
	'BigInt::sameValueZero': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-sameValueZero'
	***REMOVED***,
	'BigInt::signedRightShift': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-signedRightShift'
	***REMOVED***,
	'BigInt::subtract': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-subtract'
	***REMOVED***,
	'BigInt::toString': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-tostring'
	***REMOVED***,
	'BigInt::unaryMinus': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-unaryMinus'
	***REMOVED***,
	'BigInt::unsignedRightShift': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-bigint-unsignedRightShift'
	***REMOVED***,
	BigIntBitwiseOp: {
		url: 'https://262.ecma-international.org/12.0/#sec-bigintbitwiseop'
	***REMOVED***,
	BinaryAnd: {
		url: 'https://262.ecma-international.org/12.0/#sec-binaryand'
	***REMOVED***,
	BinaryOr: {
		url: 'https://262.ecma-international.org/12.0/#sec-binaryor'
	***REMOVED***,
	BinaryXor: {
		url: 'https://262.ecma-international.org/12.0/#sec-binaryxor'
	***REMOVED***,
	BlockDeclarationInstantiation: {
		url: 'https://262.ecma-international.org/12.0/#sec-blockdeclarationinstantiation'
	***REMOVED***,
	BoundFunctionCreate: {
		url: 'https://262.ecma-international.org/12.0/#sec-boundfunctioncreate'
	***REMOVED***,
	ByteListBitwiseOp: {
		url: 'https://262.ecma-international.org/12.0/#sec-bytelistbitwiseop'
	***REMOVED***,
	ByteListEqual: {
		url: 'https://262.ecma-international.org/12.0/#sec-bytelistequal'
	***REMOVED***,
	Call: {
		url: 'https://262.ecma-international.org/12.0/#sec-call'
	***REMOVED***,
	Canonicalize: {
		url: 'https://262.ecma-international.org/12.0/#sec-runtime-semantics-canonicalize-ch'
	***REMOVED***,
	CanonicalNumericIndexString: {
		url: 'https://262.ecma-international.org/12.0/#sec-canonicalnumericindexstring'
	***REMOVED***,
	CaseClauseIsSelected: {
		url: 'https://262.ecma-international.org/12.0/#sec-runtime-semantics-caseclauseisselected'
	***REMOVED***,
	CharacterRange: {
		url: 'https://262.ecma-international.org/12.0/#sec-runtime-semantics-characterrange-abstract-operation'
	***REMOVED***,
	CharacterRangeOrUnion: {
		url: 'https://262.ecma-international.org/12.0/#sec-runtime-semantics-characterrangeorunion-abstract-operation'
	***REMOVED***,
	CharacterSetMatcher: {
		url: 'https://262.ecma-international.org/12.0/#sec-runtime-semantics-charactersetmatcher-abstract-operation'
	***REMOVED***,
	clamp: {
		url: 'https://262.ecma-international.org/12.0/#clamping'
	***REMOVED***,
	CleanupFinalizationRegistry: {
		url: 'https://262.ecma-international.org/12.0/#sec-cleanup-finalization-registry'
	***REMOVED***,
	ClearKeptObjects: {
		url: 'https://262.ecma-international.org/12.0/#sec-clear-kept-objects'
	***REMOVED***,
	CloneArrayBuffer: {
		url: 'https://262.ecma-international.org/12.0/#sec-clonearraybuffer'
	***REMOVED***,
	CodePointAt: {
		url: 'https://262.ecma-international.org/12.0/#sec-codepointat'
	***REMOVED***,
	CodePointsToString: {
		url: 'https://262.ecma-international.org/12.0/#sec-codepointstostring'
	***REMOVED***,
	CompletePropertyDescriptor: {
		url: 'https://262.ecma-international.org/12.0/#sec-completepropertydescriptor'
	***REMOVED***,
	Completion: {
		url: 'https://262.ecma-international.org/12.0/#sec-completion-record-specification-type'
	***REMOVED***,
	CompletionRecord: {
		url: 'https://262.ecma-international.org/12.0/#sec-completion-record-specification-type'
	***REMOVED***,
	ComposeWriteEventBytes: {
		url: 'https://262.ecma-international.org/12.0/#sec-composewriteeventbytes'
	***REMOVED***,
	Construct: {
		url: 'https://262.ecma-international.org/12.0/#sec-construct'
	***REMOVED***,
	CopyDataBlockBytes: {
		url: 'https://262.ecma-international.org/12.0/#sec-copydatablockbytes'
	***REMOVED***,
	CopyDataProperties: {
		url: 'https://262.ecma-international.org/12.0/#sec-copydataproperties'
	***REMOVED***,
	CreateArrayFromList: {
		url: 'https://262.ecma-international.org/12.0/#sec-createarrayfromlist'
	***REMOVED***,
	CreateArrayIterator: {
		url: 'https://262.ecma-international.org/12.0/#sec-createarrayiterator'
	***REMOVED***,
	CreateAsyncFromSyncIterator: {
		url: 'https://262.ecma-international.org/12.0/#sec-createasyncfromsynciterator'
	***REMOVED***,
	CreateAsyncIteratorFromClosure: {
		url: 'https://262.ecma-international.org/12.0/#sec-createasynciteratorfromclosure'
	***REMOVED***,
	CreateBuiltinFunction: {
		url: 'https://262.ecma-international.org/12.0/#sec-createbuiltinfunction'
	***REMOVED***,
	CreateByteDataBlock: {
		url: 'https://262.ecma-international.org/12.0/#sec-createbytedatablock'
	***REMOVED***,
	CreateDataProperty: {
		url: 'https://262.ecma-international.org/12.0/#sec-createdataproperty'
	***REMOVED***,
	CreateDataPropertyOrThrow: {
		url: 'https://262.ecma-international.org/12.0/#sec-createdatapropertyorthrow'
	***REMOVED***,
	CreateDynamicFunction: {
		url: 'https://262.ecma-international.org/12.0/#sec-createdynamicfunction'
	***REMOVED***,
	CreateForInIterator: {
		url: 'https://262.ecma-international.org/12.0/#sec-createforiniterator'
	***REMOVED***,
	CreateHTML: {
		url: 'https://262.ecma-international.org/12.0/#sec-createhtml'
	***REMOVED***,
	CreateIntrinsics: {
		url: 'https://262.ecma-international.org/12.0/#sec-createintrinsics'
	***REMOVED***,
	CreateIteratorFromClosure: {
		url: 'https://262.ecma-international.org/12.0/#sec-createiteratorfromclosure'
	***REMOVED***,
	CreateIterResultObject: {
		url: 'https://262.ecma-international.org/12.0/#sec-createiterresultobject'
	***REMOVED***,
	CreateListFromArrayLike: {
		url: 'https://262.ecma-international.org/12.0/#sec-createlistfromarraylike'
	***REMOVED***,
	CreateListIteratorRecord: {
		url: 'https://262.ecma-international.org/12.0/#sec-createlistiteratorRecord'
	***REMOVED***,
	CreateMapIterator: {
		url: 'https://262.ecma-international.org/12.0/#sec-createmapiterator'
	***REMOVED***,
	CreateMappedArgumentsObject: {
		url: 'https://262.ecma-international.org/12.0/#sec-createmappedargumentsobject'
	***REMOVED***,
	CreateMethodProperty: {
		url: 'https://262.ecma-international.org/12.0/#sec-createmethodproperty'
	***REMOVED***,
	CreatePerIterationEnvironment: {
		url: 'https://262.ecma-international.org/12.0/#sec-createperiterationenvironment'
	***REMOVED***,
	CreateRealm: {
		url: 'https://262.ecma-international.org/12.0/#sec-createrealm'
	***REMOVED***,
	CreateRegExpStringIterator: {
		url: 'https://262.ecma-international.org/12.0/#sec-createregexpstringiterator'
	***REMOVED***,
	CreateResolvingFunctions: {
		url: 'https://262.ecma-international.org/12.0/#sec-createresolvingfunctions'
	***REMOVED***,
	CreateSetIterator: {
		url: 'https://262.ecma-international.org/12.0/#sec-createsetiterator'
	***REMOVED***,
	CreateSharedByteDataBlock: {
		url: 'https://262.ecma-international.org/12.0/#sec-createsharedbytedatablock'
	***REMOVED***,
	CreateUnmappedArgumentsObject: {
		url: 'https://262.ecma-international.org/12.0/#sec-createunmappedargumentsobject'
	***REMOVED***,
	DateFromTime: {
		url: 'https://262.ecma-international.org/12.0/#sec-date-number'
	***REMOVED***,
	DateString: {
		url: 'https://262.ecma-international.org/12.0/#sec-datestring'
	***REMOVED***,
	Day: {
		url: 'https://262.ecma-international.org/12.0/#eqn-Day'
	***REMOVED***,
	DayFromYear: {
		url: 'https://262.ecma-international.org/12.0/#eqn-DaysFromYear'
	***REMOVED***,
	DaysInYear: {
		url: 'https://262.ecma-international.org/12.0/#eqn-DaysInYear'
	***REMOVED***,
	DayWithinYear: {
		url: 'https://262.ecma-international.org/12.0/#eqn-DayWithinYear'
	***REMOVED***,
	Decode: {
		url: 'https://262.ecma-international.org/12.0/#sec-decode'
	***REMOVED***,
	DefinePropertyOrThrow: {
		url: 'https://262.ecma-international.org/12.0/#sec-definepropertyorthrow'
	***REMOVED***,
	DeletePropertyOrThrow: {
		url: 'https://262.ecma-international.org/12.0/#sec-deletepropertyorthrow'
	***REMOVED***,
	DetachArrayBuffer: {
		url: 'https://262.ecma-international.org/12.0/#sec-detacharraybuffer'
	***REMOVED***,
	Encode: {
		url: 'https://262.ecma-international.org/12.0/#sec-encode'
	***REMOVED***,
	EnterCriticalSection: {
		url: 'https://262.ecma-international.org/12.0/#sec-entercriticalsection'
	***REMOVED***,
	EnumerableOwnPropertyNames: {
		url: 'https://262.ecma-international.org/12.0/#sec-enumerableownpropertynames'
	***REMOVED***,
	EnumerateObjectProperties: {
		url: 'https://262.ecma-international.org/12.0/#sec-enumerate-object-properties'
	***REMOVED***,
	EscapeRegExpPattern: {
		url: 'https://262.ecma-international.org/12.0/#sec-escaperegexppattern'
	***REMOVED***,
	EvalDeclarationInstantiation: {
		url: 'https://262.ecma-international.org/12.0/#sec-evaldeclarationinstantiation'
	***REMOVED***,
	EvaluateCall: {
		url: 'https://262.ecma-international.org/12.0/#sec-evaluatecall'
	***REMOVED***,
	EvaluateNew: {
		url: 'https://262.ecma-international.org/12.0/#sec-evaluatenew'
	***REMOVED***,
	EvaluatePropertyAccessWithExpressionKey: {
		url: 'https://262.ecma-international.org/12.0/#sec-evaluate-property-access-with-expression-key'
	***REMOVED***,
	EvaluatePropertyAccessWithIdentifierKey: {
		url: 'https://262.ecma-international.org/12.0/#sec-evaluate-property-access-with-identifier-key'
	***REMOVED***,
	EvaluateStringOrNumericBinaryExpression: {
		url: 'https://262.ecma-international.org/12.0/#sec-evaluatestringornumericbinaryexpression'
	***REMOVED***,
	EventSet: {
		url: 'https://262.ecma-international.org/12.0/#sec-event-set'
	***REMOVED***,
	ExecuteModule: {
		url: 'https://262.ecma-international.org/12.0/#sec-source-text-module-record-execute-module'
	***REMOVED***,
	FinishDynamicImport: {
		url: 'https://262.ecma-international.org/12.0/#sec-finishdynamicimport'
	***REMOVED***,
	FlattenIntoArray: {
		url: 'https://262.ecma-international.org/12.0/#sec-flattenintoarray'
	***REMOVED***,
	floor: {
		url: 'https://262.ecma-international.org/12.0/#eqn-floor'
	***REMOVED***,
	ForBodyEvaluation: {
		url: 'https://262.ecma-international.org/12.0/#sec-forbodyevaluation'
	***REMOVED***,
	'ForIn/OfBodyEvaluation': {
		url: 'https://262.ecma-international.org/12.0/#sec-runtime-semantics-forin-div-ofbodyevaluation-lhs-stmt-iterator-lhskind-labelset'
	***REMOVED***,
	'ForIn/OfHeadEvaluation': {
		url: 'https://262.ecma-international.org/12.0/#sec-runtime-semantics-forinofheadevaluation'
	***REMOVED***,
	FromPropertyDescriptor: {
		url: 'https://262.ecma-international.org/12.0/#sec-frompropertydescriptor'
	***REMOVED***,
	FulfillPromise: {
		url: 'https://262.ecma-international.org/12.0/#sec-fulfillpromise'
	***REMOVED***,
	FunctionDeclarationInstantiation: {
		url: 'https://262.ecma-international.org/12.0/#sec-functiondeclarationinstantiation'
	***REMOVED***,
	GeneratorResume: {
		url: 'https://262.ecma-international.org/12.0/#sec-generatorresume'
	***REMOVED***,
	GeneratorResumeAbrupt: {
		url: 'https://262.ecma-international.org/12.0/#sec-generatorresumeabrupt'
	***REMOVED***,
	GeneratorStart: {
		url: 'https://262.ecma-international.org/12.0/#sec-generatorstart'
	***REMOVED***,
	GeneratorValidate: {
		url: 'https://262.ecma-international.org/12.0/#sec-generatorvalidate'
	***REMOVED***,
	GeneratorYield: {
		url: 'https://262.ecma-international.org/12.0/#sec-generatoryield'
	***REMOVED***,
	Get: {
		url: 'https://262.ecma-international.org/12.0/#sec-get-o-p'
	***REMOVED***,
	GetActiveScriptOrModule: {
		url: 'https://262.ecma-international.org/12.0/#sec-getactivescriptormodule'
	***REMOVED***,
	GetFunctionRealm: {
		url: 'https://262.ecma-international.org/12.0/#sec-getfunctionrealm'
	***REMOVED***,
	GetGeneratorKind: {
		url: 'https://262.ecma-international.org/12.0/#sec-getgeneratorkind'
	***REMOVED***,
	GetGlobalObject: {
		url: 'https://262.ecma-international.org/12.0/#sec-getglobalobject'
	***REMOVED***,
	GetIdentifierReference: {
		url: 'https://262.ecma-international.org/12.0/#sec-getidentifierreference'
	***REMOVED***,
	GetIterator: {
		url: 'https://262.ecma-international.org/12.0/#sec-getiterator'
	***REMOVED***,
	GetMethod: {
		url: 'https://262.ecma-international.org/12.0/#sec-getmethod'
	***REMOVED***,
	GetModifySetValueInBuffer: {
		url: 'https://262.ecma-international.org/12.0/#sec-getmodifysetvalueinbuffer'
	***REMOVED***,
	GetModuleNamespace: {
		url: 'https://262.ecma-international.org/12.0/#sec-getmodulenamespace'
	***REMOVED***,
	GetNewTarget: {
		url: 'https://262.ecma-international.org/12.0/#sec-getnewtarget'
	***REMOVED***,
	GetOwnPropertyKeys: {
		url: 'https://262.ecma-international.org/12.0/#sec-getownpropertykeys'
	***REMOVED***,
	GetPromiseResolve: {
		url: 'https://262.ecma-international.org/12.0/#sec-getpromiseresolve'
	***REMOVED***,
	GetPrototypeFromConstructor: {
		url: 'https://262.ecma-international.org/12.0/#sec-getprototypefromconstructor'
	***REMOVED***,
	GetSubstitution: {
		url: 'https://262.ecma-international.org/12.0/#sec-getsubstitution'
	***REMOVED***,
	GetSuperConstructor: {
		url: 'https://262.ecma-international.org/12.0/#sec-getsuperconstructor'
	***REMOVED***,
	GetTemplateObject: {
		url: 'https://262.ecma-international.org/12.0/#sec-gettemplateobject'
	***REMOVED***,
	GetThisEnvironment: {
		url: 'https://262.ecma-international.org/12.0/#sec-getthisenvironment'
	***REMOVED***,
	GetThisValue: {
		url: 'https://262.ecma-international.org/12.0/#sec-getthisvalue'
	***REMOVED***,
	GetV: {
		url: 'https://262.ecma-international.org/12.0/#sec-getv'
	***REMOVED***,
	GetValue: {
		url: 'https://262.ecma-international.org/12.0/#sec-getvalue'
	***REMOVED***,
	GetValueFromBuffer: {
		url: 'https://262.ecma-international.org/12.0/#sec-getvaluefrombuffer'
	***REMOVED***,
	GetViewValue: {
		url: 'https://262.ecma-international.org/12.0/#sec-getviewvalue'
	***REMOVED***,
	GetWaiterList: {
		url: 'https://262.ecma-international.org/12.0/#sec-getwaiterlist'
	***REMOVED***,
	GlobalDeclarationInstantiation: {
		url: 'https://262.ecma-international.org/12.0/#sec-globaldeclarationinstantiation'
	***REMOVED***,
	'happens-before': {
		url: 'https://262.ecma-international.org/12.0/#sec-happens-before'
	***REMOVED***,
	HasOwnProperty: {
		url: 'https://262.ecma-international.org/12.0/#sec-hasownproperty'
	***REMOVED***,
	HasProperty: {
		url: 'https://262.ecma-international.org/12.0/#sec-hasproperty'
	***REMOVED***,
	'host-synchronizes-with': {
		url: 'https://262.ecma-international.org/12.0/#sec-host-synchronizes-with'
	***REMOVED***,
	HostEventSet: {
		url: 'https://262.ecma-international.org/12.0/#sec-hosteventset'
	***REMOVED***,
	HourFromTime: {
		url: 'https://262.ecma-international.org/12.0/#eqn-HourFromTime'
	***REMOVED***,
	IfAbruptRejectPromise: {
		url: 'https://262.ecma-international.org/12.0/#sec-ifabruptrejectpromise'
	***REMOVED***,
	ImportedLocalNames: {
		url: 'https://262.ecma-international.org/12.0/#sec-importedlocalnames'
	***REMOVED***,
	InitializeBoundName: {
		url: 'https://262.ecma-international.org/12.0/#sec-initializeboundname'
	***REMOVED***,
	InitializeEnvironment: {
		url: 'https://262.ecma-international.org/12.0/#sec-source-text-module-record-initialize-environment'
	***REMOVED***,
	InitializeHostDefinedRealm: {
		url: 'https://262.ecma-international.org/12.0/#sec-initializehostdefinedrealm'
	***REMOVED***,
	InitializeReferencedBinding: {
		url: 'https://262.ecma-international.org/12.0/#sec-initializereferencedbinding'
	***REMOVED***,
	InitializeTypedArrayFromArrayBuffer: {
		url: 'https://262.ecma-international.org/12.0/#sec-initializetypedarrayfromarraybuffer'
	***REMOVED***,
	InitializeTypedArrayFromArrayLike: {
		url: 'https://262.ecma-international.org/12.0/#sec-initializetypedarrayfromarraylike'
	***REMOVED***,
	InitializeTypedArrayFromList: {
		url: 'https://262.ecma-international.org/12.0/#sec-initializetypedarrayfromlist'
	***REMOVED***,
	InitializeTypedArrayFromTypedArray: {
		url: 'https://262.ecma-international.org/12.0/#sec-initializetypedarrayfromtypedarray'
	***REMOVED***,
	InLeapYear: {
		url: 'https://262.ecma-international.org/12.0/#eqn-InLeapYear'
	***REMOVED***,
	InnerModuleEvaluation: {
		url: 'https://262.ecma-international.org/12.0/#sec-innermoduleevaluation'
	***REMOVED***,
	InnerModuleLinking: {
		url: 'https://262.ecma-international.org/12.0/#sec-InnerModuleLinking'
	***REMOVED***,
	InstanceofOperator: {
		url: 'https://262.ecma-international.org/12.0/#sec-instanceofoperator'
	***REMOVED***,
	IntegerIndexedElementGet: {
		url: 'https://262.ecma-international.org/12.0/#sec-integerindexedelementget'
	***REMOVED***,
	IntegerIndexedElementSet: {
		url: 'https://262.ecma-international.org/12.0/#sec-integerindexedelementset'
	***REMOVED***,
	IntegerIndexedObjectCreate: {
		url: 'https://262.ecma-international.org/12.0/#sec-integerindexedobjectcreate'
	***REMOVED***,
	InternalizeJSONProperty: {
		url: 'https://262.ecma-international.org/12.0/#sec-internalizejsonproperty'
	***REMOVED***,
	Invoke: {
		url: 'https://262.ecma-international.org/12.0/#sec-invoke'
	***REMOVED***,
	IsAccessorDescriptor: {
		url: 'https://262.ecma-international.org/12.0/#sec-isaccessordescriptor'
	***REMOVED***,
	IsAnonymousFunctionDefinition: {
		url: 'https://262.ecma-international.org/12.0/#sec-isanonymousfunctiondefinition'
	***REMOVED***,
	IsArray: {
		url: 'https://262.ecma-international.org/12.0/#sec-isarray'
	***REMOVED***,
	IsBigIntElementType: {
		url: 'https://262.ecma-international.org/12.0/#sec-isbigintelementtype'
	***REMOVED***,
	IsCallable: {
		url: 'https://262.ecma-international.org/12.0/#sec-iscallable'
	***REMOVED***,
	IsCompatiblePropertyDescriptor: {
		url: 'https://262.ecma-international.org/12.0/#sec-iscompatiblepropertydescriptor'
	***REMOVED***,
	IsConcatSpreadable: {
		url: 'https://262.ecma-international.org/12.0/#sec-isconcatspreadable'
	***REMOVED***,
	IsConstructor: {
		url: 'https://262.ecma-international.org/12.0/#sec-isconstructor'
	***REMOVED***,
	IsDataDescriptor: {
		url: 'https://262.ecma-international.org/12.0/#sec-isdatadescriptor'
	***REMOVED***,
	IsDetachedBuffer: {
		url: 'https://262.ecma-international.org/12.0/#sec-isdetachedbuffer'
	***REMOVED***,
	IsExtensible: {
		url: 'https://262.ecma-international.org/12.0/#sec-isextensible-o'
	***REMOVED***,
	IsGenericDescriptor: {
		url: 'https://262.ecma-international.org/12.0/#sec-isgenericdescriptor'
	***REMOVED***,
	IsInTailPosition: {
		url: 'https://262.ecma-international.org/12.0/#sec-isintailposition'
	***REMOVED***,
	IsIntegralNumber: {
		url: 'https://262.ecma-international.org/12.0/#sec-isintegralnumber'
	***REMOVED***,
	IsLabelledFunction: {
		url: 'https://262.ecma-international.org/12.0/#sec-islabelledfunction'
	***REMOVED***,
	IsNoTearConfiguration: {
		url: 'https://262.ecma-international.org/12.0/#sec-isnotearconfiguration'
	***REMOVED***,
	IsPromise: {
		url: 'https://262.ecma-international.org/12.0/#sec-ispromise'
	***REMOVED***,
	IsPropertyKey: {
		url: 'https://262.ecma-international.org/12.0/#sec-ispropertykey'
	***REMOVED***,
	IsPropertyReference: {
		url: 'https://262.ecma-international.org/12.0/#sec-ispropertyreference'
	***REMOVED***,
	IsRegExp: {
		url: 'https://262.ecma-international.org/12.0/#sec-isregexp'
	***REMOVED***,
	IsSharedArrayBuffer: {
		url: 'https://262.ecma-international.org/12.0/#sec-issharedarraybuffer'
	***REMOVED***,
	IsStringPrefix: {
		url: 'https://262.ecma-international.org/12.0/#sec-isstringprefix'
	***REMOVED***,
	IsSuperReference: {
		url: 'https://262.ecma-international.org/12.0/#sec-issuperreference'
	***REMOVED***,
	IsUnclampedIntegerElementType: {
		url: 'https://262.ecma-international.org/12.0/#sec-isunclampedintegerelementtype'
	***REMOVED***,
	IsUnresolvableReference: {
		url: 'https://262.ecma-international.org/12.0/#sec-isunresolvablereference'
	***REMOVED***,
	IsUnsignedElementType: {
		url: 'https://262.ecma-international.org/12.0/#sec-isunsignedelementtype'
	***REMOVED***,
	IsValidIntegerIndex: {
		url: 'https://262.ecma-international.org/12.0/#sec-isvalidintegerindex'
	***REMOVED***,
	IsValidRegularExpressionLiteral: {
		url: 'https://262.ecma-international.org/12.0/#sec-isvalidregularexpressionliteral'
	***REMOVED***,
	IsWordChar: {
		url: 'https://262.ecma-international.org/12.0/#sec-runtime-semantics-iswordchar-abstract-operation'
	***REMOVED***,
	IterableToList: {
		url: 'https://262.ecma-international.org/12.0/#sec-iterabletolist'
	***REMOVED***,
	IteratorClose: {
		url: 'https://262.ecma-international.org/12.0/#sec-iteratorclose'
	***REMOVED***,
	IteratorComplete: {
		url: 'https://262.ecma-international.org/12.0/#sec-iteratorcomplete'
	***REMOVED***,
	IteratorNext: {
		url: 'https://262.ecma-international.org/12.0/#sec-iteratornext'
	***REMOVED***,
	IteratorStep: {
		url: 'https://262.ecma-international.org/12.0/#sec-iteratorstep'
	***REMOVED***,
	IteratorValue: {
		url: 'https://262.ecma-international.org/12.0/#sec-iteratorvalue'
	***REMOVED***,
	LeaveCriticalSection: {
		url: 'https://262.ecma-international.org/12.0/#sec-leavecriticalsection'
	***REMOVED***,
	LengthOfArrayLike: {
		url: 'https://262.ecma-international.org/12.0/#sec-lengthofarraylike'
	***REMOVED***,
	LocalTime: {
		url: 'https://262.ecma-international.org/12.0/#sec-localtime'
	***REMOVED***,
	LocalTZA: {
		url: 'https://262.ecma-international.org/12.0/#sec-local-time-zone-adjustment'
	***REMOVED***,
	LoopContinues: {
		url: 'https://262.ecma-international.org/12.0/#sec-loopcontinues'
	***REMOVED***,
	MakeArgGetter: {
		url: 'https://262.ecma-international.org/12.0/#sec-makearggetter'
	***REMOVED***,
	MakeArgSetter: {
		url: 'https://262.ecma-international.org/12.0/#sec-makeargsetter'
	***REMOVED***,
	MakeBasicObject: {
		url: 'https://262.ecma-international.org/12.0/#sec-makebasicobject'
	***REMOVED***,
	MakeClassConstructor: {
		url: 'https://262.ecma-international.org/12.0/#sec-makeclassconstructor'
	***REMOVED***,
	MakeConstructor: {
		url: 'https://262.ecma-international.org/12.0/#sec-makeconstructor'
	***REMOVED***,
	MakeDate: {
		url: 'https://262.ecma-international.org/12.0/#sec-makedate'
	***REMOVED***,
	MakeDay: {
		url: 'https://262.ecma-international.org/12.0/#sec-makeday'
	***REMOVED***,
	MakeMethod: {
		url: 'https://262.ecma-international.org/12.0/#sec-makemethod'
	***REMOVED***,
	MakeSuperPropertyReference: {
		url: 'https://262.ecma-international.org/12.0/#sec-makesuperpropertyreference'
	***REMOVED***,
	MakeTime: {
		url: 'https://262.ecma-international.org/12.0/#sec-maketime'
	***REMOVED***,
	max: {
		url: 'https://262.ecma-international.org/12.0/#eqn-max'
	***REMOVED***,
	'memory-order': {
		url: 'https://262.ecma-international.org/12.0/#sec-memory-order'
	***REMOVED***,
	min: {
		url: 'https://262.ecma-international.org/12.0/#eqn-min'
	***REMOVED***,
	MinFromTime: {
		url: 'https://262.ecma-international.org/12.0/#eqn-MinFromTime'
	***REMOVED***,
	ModuleNamespaceCreate: {
		url: 'https://262.ecma-international.org/12.0/#sec-modulenamespacecreate'
	***REMOVED***,
	modulo: {
		url: 'https://262.ecma-international.org/12.0/#eqn-modulo'
	***REMOVED***,
	MonthFromTime: {
		url: 'https://262.ecma-international.org/12.0/#eqn-MonthFromTime'
	***REMOVED***,
	msFromTime: {
		url: 'https://262.ecma-international.org/12.0/#eqn-msFromTime'
	***REMOVED***,
	NewDeclarativeEnvironment: {
		url: 'https://262.ecma-international.org/12.0/#sec-newdeclarativeenvironment'
	***REMOVED***,
	NewFunctionEnvironment: {
		url: 'https://262.ecma-international.org/12.0/#sec-newfunctionenvironment'
	***REMOVED***,
	NewGlobalEnvironment: {
		url: 'https://262.ecma-international.org/12.0/#sec-newglobalenvironment'
	***REMOVED***,
	NewModuleEnvironment: {
		url: 'https://262.ecma-international.org/12.0/#sec-newmoduleenvironment'
	***REMOVED***,
	NewObjectEnvironment: {
		url: 'https://262.ecma-international.org/12.0/#sec-newobjectenvironment'
	***REMOVED***,
	NewPromiseCapability: {
		url: 'https://262.ecma-international.org/12.0/#sec-newpromisecapability'
	***REMOVED***,
	NewPromiseReactionJob: {
		url: 'https://262.ecma-international.org/12.0/#sec-newpromisereactionjob'
	***REMOVED***,
	NewPromiseResolveThenableJob: {
		url: 'https://262.ecma-international.org/12.0/#sec-newpromiseresolvethenablejob'
	***REMOVED***,
	NormalCompletion: {
		url: 'https://262.ecma-international.org/12.0/#sec-normalcompletion'
	***REMOVED***,
	NotifyWaiter: {
		url: 'https://262.ecma-international.org/12.0/#sec-notifywaiter'
	***REMOVED***,
	'Number::add': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-add'
	***REMOVED***,
	'Number::bitwiseAND': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-bitwiseAND'
	***REMOVED***,
	'Number::bitwiseNOT': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-bitwiseNOT'
	***REMOVED***,
	'Number::bitwiseOR': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-bitwiseOR'
	***REMOVED***,
	'Number::bitwiseXOR': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-bitwiseXOR'
	***REMOVED***,
	'Number::divide': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-divide'
	***REMOVED***,
	'Number::equal': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-equal'
	***REMOVED***,
	'Number::exponentiate': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-exponentiate'
	***REMOVED***,
	'Number::leftShift': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-leftShift'
	***REMOVED***,
	'Number::lessThan': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-lessThan'
	***REMOVED***,
	'Number::multiply': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-multiply'
	***REMOVED***,
	'Number::remainder': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-remainder'
	***REMOVED***,
	'Number::sameValue': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-sameValue'
	***REMOVED***,
	'Number::sameValueZero': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-sameValueZero'
	***REMOVED***,
	'Number::signedRightShift': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-signedRightShift'
	***REMOVED***,
	'Number::subtract': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-subtract'
	***REMOVED***,
	'Number::toString': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-tostring'
	***REMOVED***,
	'Number::unaryMinus': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-unaryMinus'
	***REMOVED***,
	'Number::unsignedRightShift': {
		url: 'https://262.ecma-international.org/12.0/#sec-numeric-types-number-unsignedRightShift'
	***REMOVED***,
	NumberBitwiseOp: {
		url: 'https://262.ecma-international.org/12.0/#sec-numberbitwiseop'
	***REMOVED***,
	NumberToBigInt: {
		url: 'https://262.ecma-international.org/12.0/#sec-numbertobigint'
	***REMOVED***,
	NumericToRawBytes: {
		url: 'https://262.ecma-international.org/12.0/#sec-numerictorawbytes'
	***REMOVED***,
	ObjectDefineProperties: {
		url: 'https://262.ecma-international.org/12.0/#sec-objectdefineproperties'
	***REMOVED***,
	OrdinaryCallBindThis: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinarycallbindthis'
	***REMOVED***,
	OrdinaryCallEvaluateBody: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinarycallevaluatebody'
	***REMOVED***,
	OrdinaryCreateFromConstructor: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinarycreatefromconstructor'
	***REMOVED***,
	OrdinaryDefineOwnProperty: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinarydefineownproperty'
	***REMOVED***,
	OrdinaryDelete: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinarydelete'
	***REMOVED***,
	OrdinaryFunctionCreate: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinaryfunctioncreate'
	***REMOVED***,
	OrdinaryGet: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinaryget'
	***REMOVED***,
	OrdinaryGetOwnProperty: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinarygetownproperty'
	***REMOVED***,
	OrdinaryGetPrototypeOf: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinarygetprototypeof'
	***REMOVED***,
	OrdinaryHasInstance: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinaryhasinstance'
	***REMOVED***,
	OrdinaryHasProperty: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinaryhasproperty'
	***REMOVED***,
	OrdinaryIsExtensible: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinaryisextensible'
	***REMOVED***,
	OrdinaryObjectCreate: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinaryobjectcreate'
	***REMOVED***,
	OrdinaryOwnPropertyKeys: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinaryownpropertykeys'
	***REMOVED***,
	OrdinaryPreventExtensions: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinarypreventextensions'
	***REMOVED***,
	OrdinarySet: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinaryset'
	***REMOVED***,
	OrdinarySetPrototypeOf: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinarysetprototypeof'
	***REMOVED***,
	OrdinarySetWithOwnDescriptor: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinarysetwithowndescriptor'
	***REMOVED***,
	OrdinaryToPrimitive: {
		url: 'https://262.ecma-international.org/12.0/#sec-ordinarytoprimitive'
	***REMOVED***,
	ParseModule: {
		url: 'https://262.ecma-international.org/12.0/#sec-parsemodule'
	***REMOVED***,
	ParsePattern: {
		url: 'https://262.ecma-international.org/12.0/#sec-parsepattern'
	***REMOVED***,
	ParseScript: {
		url: 'https://262.ecma-international.org/12.0/#sec-parse-script'
	***REMOVED***,
	ParseText: {
		url: 'https://262.ecma-international.org/12.0/#sec-parsetext'
	***REMOVED***,
	PerformEval: {
		url: 'https://262.ecma-international.org/12.0/#sec-performeval'
	***REMOVED***,
	PerformPromiseAll: {
		url: 'https://262.ecma-international.org/12.0/#sec-performpromiseall'
	***REMOVED***,
	PerformPromiseAllSettled: {
		url: 'https://262.ecma-international.org/12.0/#sec-performpromiseallsettled'
	***REMOVED***,
	PerformPromiseAny: {
		url: 'https://262.ecma-international.org/12.0/#sec-performpromiseany'
	***REMOVED***,
	PerformPromiseRace: {
		url: 'https://262.ecma-international.org/12.0/#sec-performpromiserace'
	***REMOVED***,
	PerformPromiseThen: {
		url: 'https://262.ecma-international.org/12.0/#sec-performpromisethen'
	***REMOVED***,
	PrepareForOrdinaryCall: {
		url: 'https://262.ecma-international.org/12.0/#sec-prepareforordinarycall'
	***REMOVED***,
	PrepareForTailCall: {
		url: 'https://262.ecma-international.org/12.0/#sec-preparefortailcall'
	***REMOVED***,
	PromiseResolve: {
		url: 'https://262.ecma-international.org/12.0/#sec-promise-resolve'
	***REMOVED***,
	ProxyCreate: {
		url: 'https://262.ecma-international.org/12.0/#sec-proxycreate'
	***REMOVED***,
	PutValue: {
		url: 'https://262.ecma-international.org/12.0/#sec-putvalue'
	***REMOVED***,
	QuoteJSONString: {
		url: 'https://262.ecma-international.org/12.0/#sec-quotejsonstring'
	***REMOVED***,
	RawBytesToNumeric: {
		url: 'https://262.ecma-international.org/12.0/#sec-rawbytestonumeric'
	***REMOVED***,
	'reads-bytes-from': {
		url: 'https://262.ecma-international.org/12.0/#sec-reads-bytes-from'
	***REMOVED***,
	'reads-from': {
		url: 'https://262.ecma-international.org/12.0/#sec-reads-from'
	***REMOVED***,
	RegExpAlloc: {
		url: 'https://262.ecma-international.org/12.0/#sec-regexpalloc'
	***REMOVED***,
	RegExpBuiltinExec: {
		url: 'https://262.ecma-international.org/12.0/#sec-regexpbuiltinexec'
	***REMOVED***,
	RegExpCreate: {
		url: 'https://262.ecma-international.org/12.0/#sec-regexpcreate'
	***REMOVED***,
	RegExpExec: {
		url: 'https://262.ecma-international.org/12.0/#sec-regexpexec'
	***REMOVED***,
	RegExpInitialize: {
		url: 'https://262.ecma-international.org/12.0/#sec-regexpinitialize'
	***REMOVED***,
	RejectPromise: {
		url: 'https://262.ecma-international.org/12.0/#sec-rejectpromise'
	***REMOVED***,
	RemoveWaiter: {
		url: 'https://262.ecma-international.org/12.0/#sec-removewaiter'
	***REMOVED***,
	RemoveWaiters: {
		url: 'https://262.ecma-international.org/12.0/#sec-removewaiters'
	***REMOVED***,
	RepeatMatcher: {
		url: 'https://262.ecma-international.org/12.0/#sec-runtime-semantics-repeatmatcher-abstract-operation'
	***REMOVED***,
	RequireInternalSlot: {
		url: 'https://262.ecma-international.org/12.0/#sec-requireinternalslot'
	***REMOVED***,
	RequireObjectCoercible: {
		url: 'https://262.ecma-international.org/12.0/#sec-requireobjectcoercible'
	***REMOVED***,
	ResolveBinding: {
		url: 'https://262.ecma-international.org/12.0/#sec-resolvebinding'
	***REMOVED***,
	ResolveThisBinding: {
		url: 'https://262.ecma-international.org/12.0/#sec-resolvethisbinding'
	***REMOVED***,
	ReturnIfAbrupt: {
		url: 'https://262.ecma-international.org/12.0/#sec-returnifabrupt'
	***REMOVED***,
	SameValue: {
		url: 'https://262.ecma-international.org/12.0/#sec-samevalue'
	***REMOVED***,
	SameValueNonNumeric: {
		url: 'https://262.ecma-international.org/12.0/#sec-samevaluenonnumeric'
	***REMOVED***,
	SameValueZero: {
		url: 'https://262.ecma-international.org/12.0/#sec-samevaluezero'
	***REMOVED***,
	ScriptEvaluation: {
		url: 'https://262.ecma-international.org/12.0/#sec-runtime-semantics-scriptevaluation'
	***REMOVED***,
	SecFromTime: {
		url: 'https://262.ecma-international.org/12.0/#eqn-SecFromTime'
	***REMOVED***,
	SerializeJSONArray: {
		url: 'https://262.ecma-international.org/12.0/#sec-serializejsonarray'
	***REMOVED***,
	SerializeJSONObject: {
		url: 'https://262.ecma-international.org/12.0/#sec-serializejsonobject'
	***REMOVED***,
	SerializeJSONProperty: {
		url: 'https://262.ecma-international.org/12.0/#sec-serializejsonproperty'
	***REMOVED***,
	Set: {
		url: 'https://262.ecma-international.org/12.0/#sec-set-o-p-v-throw'
	***REMOVED***,
	SetDefaultGlobalBindings: {
		url: 'https://262.ecma-international.org/12.0/#sec-setdefaultglobalbindings'
	***REMOVED***,
	SetFunctionLength: {
		url: 'https://262.ecma-international.org/12.0/#sec-setfunctionlength'
	***REMOVED***,
	SetFunctionName: {
		url: 'https://262.ecma-international.org/12.0/#sec-setfunctionname'
	***REMOVED***,
	SetImmutablePrototype: {
		url: 'https://262.ecma-international.org/12.0/#sec-set-immutable-prototype'
	***REMOVED***,
	SetIntegrityLevel: {
		url: 'https://262.ecma-international.org/12.0/#sec-setintegritylevel'
	***REMOVED***,
	SetRealmGlobalObject: {
		url: 'https://262.ecma-international.org/12.0/#sec-setrealmglobalobject'
	***REMOVED***,
	SetTypedArrayFromArrayLike: {
		url: 'https://262.ecma-international.org/12.0/#sec-settypedarrayfromarraylike'
	***REMOVED***,
	SetTypedArrayFromTypedArray: {
		url: 'https://262.ecma-international.org/12.0/#sec-settypedarrayfromtypedarray'
	***REMOVED***,
	SetValueInBuffer: {
		url: 'https://262.ecma-international.org/12.0/#sec-setvalueinbuffer'
	***REMOVED***,
	SetViewValue: {
		url: 'https://262.ecma-international.org/12.0/#sec-setviewvalue'
	***REMOVED***,
	SharedDataBlockEventSet: {
		url: 'https://262.ecma-international.org/12.0/#sec-sharedatablockeventset'
	***REMOVED***,
	SortCompare: {
		url: 'https://262.ecma-international.org/12.0/#sec-sortcompare'
	***REMOVED***,
	SpeciesConstructor: {
		url: 'https://262.ecma-international.org/12.0/#sec-speciesconstructor'
	***REMOVED***,
	SplitMatch: {
		url: 'https://262.ecma-international.org/12.0/#sec-splitmatch'
	***REMOVED***,
	'Strict Equality Comparison': {
		url: 'https://262.ecma-international.org/12.0/#sec-strict-equality-comparison'
	***REMOVED***,
	StringCreate: {
		url: 'https://262.ecma-international.org/12.0/#sec-stringcreate'
	***REMOVED***,
	StringGetOwnProperty: {
		url: 'https://262.ecma-international.org/12.0/#sec-stringgetownproperty'
	***REMOVED***,
	StringIndexOf: {
		url: 'https://262.ecma-international.org/12.0/#sec-stringindexof'
	***REMOVED***,
	StringPad: {
		url: 'https://262.ecma-international.org/12.0/#sec-stringpad'
	***REMOVED***,
	StringToBigInt: {
		url: 'https://262.ecma-international.org/12.0/#sec-stringtobigint'
	***REMOVED***,
	StringToCodePoints: {
		url: 'https://262.ecma-international.org/12.0/#sec-stringtocodepoints'
	***REMOVED***,
	substring: {
		url: 'https://262.ecma-international.org/12.0/#substring'
	***REMOVED***,
	SuspendAgent: {
		url: 'https://262.ecma-international.org/12.0/#sec-suspendagent'
	***REMOVED***,
	SymbolDescriptiveString: {
		url: 'https://262.ecma-international.org/12.0/#sec-symboldescriptivestring'
	***REMOVED***,
	'synchronizes-with': {
		url: 'https://262.ecma-international.org/12.0/#sec-synchronizes-with'
	***REMOVED***,
	TestIntegrityLevel: {
		url: 'https://262.ecma-international.org/12.0/#sec-testintegritylevel'
	***REMOVED***,
	thisBigIntValue: {
		url: 'https://262.ecma-international.org/12.0/#thisbigintvalue'
	***REMOVED***,
	thisBooleanValue: {
		url: 'https://262.ecma-international.org/12.0/#thisbooleanvalue'
	***REMOVED***,
	thisNumberValue: {
		url: 'https://262.ecma-international.org/12.0/#thisnumbervalue'
	***REMOVED***,
	thisStringValue: {
		url: 'https://262.ecma-international.org/12.0/#thisstringvalue'
	***REMOVED***,
	thisSymbolValue: {
		url: 'https://262.ecma-international.org/12.0/#thissymbolvalue'
	***REMOVED***,
	thisTimeValue: {
		url: 'https://262.ecma-international.org/12.0/#thistimevalue'
	***REMOVED***,
	ThrowCompletion: {
		url: 'https://262.ecma-international.org/12.0/#sec-throwcompletion'
	***REMOVED***,
	TimeClip: {
		url: 'https://262.ecma-international.org/12.0/#sec-timeclip'
	***REMOVED***,
	TimeFromYear: {
		url: 'https://262.ecma-international.org/12.0/#eqn-TimeFromYear'
	***REMOVED***,
	TimeString: {
		url: 'https://262.ecma-international.org/12.0/#sec-timestring'
	***REMOVED***,
	TimeWithinDay: {
		url: 'https://262.ecma-international.org/12.0/#eqn-TimeWithinDay'
	***REMOVED***,
	TimeZoneString: {
		url: 'https://262.ecma-international.org/12.0/#sec-timezoneestring'
	***REMOVED***,
	ToBigInt: {
		url: 'https://262.ecma-international.org/12.0/#sec-tobigint'
	***REMOVED***,
	ToBigInt64: {
		url: 'https://262.ecma-international.org/12.0/#sec-tobigint64'
	***REMOVED***,
	ToBigUint64: {
		url: 'https://262.ecma-international.org/12.0/#sec-tobiguint64'
	***REMOVED***,
	ToBoolean: {
		url: 'https://262.ecma-international.org/12.0/#sec-toboolean'
	***REMOVED***,
	ToDateString: {
		url: 'https://262.ecma-international.org/12.0/#sec-todatestring'
	***REMOVED***,
	ToIndex: {
		url: 'https://262.ecma-international.org/12.0/#sec-toindex'
	***REMOVED***,
	ToInt16: {
		url: 'https://262.ecma-international.org/12.0/#sec-toint16'
	***REMOVED***,
	ToInt32: {
		url: 'https://262.ecma-international.org/12.0/#sec-toint32'
	***REMOVED***,
	ToInt8: {
		url: 'https://262.ecma-international.org/12.0/#sec-toint8'
	***REMOVED***,
	ToIntegerOrInfinity: {
		url: 'https://262.ecma-international.org/12.0/#sec-tointegerorinfinity'
	***REMOVED***,
	ToLength: {
		url: 'https://262.ecma-international.org/12.0/#sec-tolength'
	***REMOVED***,
	ToNumber: {
		url: 'https://262.ecma-international.org/12.0/#sec-tonumber'
	***REMOVED***,
	ToNumeric: {
		url: 'https://262.ecma-international.org/12.0/#sec-tonumeric'
	***REMOVED***,
	ToObject: {
		url: 'https://262.ecma-international.org/12.0/#sec-toobject'
	***REMOVED***,
	ToPrimitive: {
		url: 'https://262.ecma-international.org/12.0/#sec-toprimitive'
	***REMOVED***,
	ToPropertyDescriptor: {
		url: 'https://262.ecma-international.org/12.0/#sec-topropertydescriptor'
	***REMOVED***,
	ToPropertyKey: {
		url: 'https://262.ecma-international.org/12.0/#sec-topropertykey'
	***REMOVED***,
	ToString: {
		url: 'https://262.ecma-international.org/12.0/#sec-tostring'
	***REMOVED***,
	ToUint16: {
		url: 'https://262.ecma-international.org/12.0/#sec-touint16'
	***REMOVED***,
	ToUint32: {
		url: 'https://262.ecma-international.org/12.0/#sec-touint32'
	***REMOVED***,
	ToUint8: {
		url: 'https://262.ecma-international.org/12.0/#sec-touint8'
	***REMOVED***,
	ToUint8Clamp: {
		url: 'https://262.ecma-international.org/12.0/#sec-touint8clamp'
	***REMOVED***,
	TriggerPromiseReactions: {
		url: 'https://262.ecma-international.org/12.0/#sec-triggerpromisereactions'
	***REMOVED***,
	TrimString: {
		url: 'https://262.ecma-international.org/12.0/#sec-trimstring'
	***REMOVED***,
	Type: {
		url: 'https://262.ecma-international.org/12.0/#sec-ecmascript-data-types-and-values'
	***REMOVED***,
	TypedArrayCreate: {
		url: 'https://262.ecma-international.org/12.0/#typedarray-create'
	***REMOVED***,
	TypedArraySpeciesCreate: {
		url: 'https://262.ecma-international.org/12.0/#typedarray-species-create'
	***REMOVED***,
	UnicodeEscape: {
		url: 'https://262.ecma-international.org/12.0/#sec-unicodeescape'
	***REMOVED***,
	UnicodeMatchProperty: {
		url: 'https://262.ecma-international.org/12.0/#sec-runtime-semantics-unicodematchproperty-p'
	***REMOVED***,
	UnicodeMatchPropertyValue: {
		url: 'https://262.ecma-international.org/12.0/#sec-runtime-semantics-unicodematchpropertyvalue-p-v'
	***REMOVED***,
	UpdateEmpty: {
		url: 'https://262.ecma-international.org/12.0/#sec-updateempty'
	***REMOVED***,
	UTC: {
		url: 'https://262.ecma-international.org/12.0/#sec-utc-t'
	***REMOVED***,
	UTF16EncodeCodePoint: {
		url: 'https://262.ecma-international.org/12.0/#sec-utf16encodecodepoint'
	***REMOVED***,
	UTF16SurrogatePairToCodePoint: {
		url: 'https://262.ecma-international.org/12.0/#sec-utf16decodesurrogatepair'
	***REMOVED***,
	ValidateAndApplyPropertyDescriptor: {
		url: 'https://262.ecma-international.org/12.0/#sec-validateandapplypropertydescriptor'
	***REMOVED***,
	ValidateAtomicAccess: {
		url: 'https://262.ecma-international.org/12.0/#sec-validateatomicaccess'
	***REMOVED***,
	ValidateIntegerTypedArray: {
		url: 'https://262.ecma-international.org/12.0/#sec-validateintegertypedarray'
	***REMOVED***,
	ValidateTypedArray: {
		url: 'https://262.ecma-international.org/12.0/#sec-validatetypedarray'
	***REMOVED***,
	ValueOfReadEvent: {
		url: 'https://262.ecma-international.org/12.0/#sec-valueofreadevent'
	***REMOVED***,
	WeakRefDeref: {
		url: 'https://262.ecma-international.org/12.0/#sec-weakrefderef'
	***REMOVED***,
	WeekDay: {
		url: 'https://262.ecma-international.org/12.0/#sec-week-day'
	***REMOVED***,
	WordCharacters: {
		url: 'https://262.ecma-international.org/12.0/#sec-runtime-semantics-wordcharacters-abstract-operation'
	***REMOVED***,
	YearFromTime: {
		url: 'https://262.ecma-international.org/12.0/#eqn-YearFromTime'
	***REMOVED***,
	Yield: {
		url: 'https://262.ecma-international.org/12.0/#sec-yield'
	***REMOVED***
***REMOVED***;
