// test-exclude/index.js
'use strict';

const path = require('path');
const { promisify ***REMOVED*** = require('util');
const glob = promisify(require('glob'));
const minimatch = require('minimatch');
const { defaults ***REMOVED*** = require('@istanbuljs/schema');
const isOutsideDir = require('./is-outside-dir');

class TestExclude {
    constructor(opts = {***REMOVED***) {
        Object.assign(
            this,
            {relativePath: true***REMOVED***,
            defaults.testExclude
        );

        for (const [name, value] of Object.entries(opts)) {
            if (value !== undefined) {
                this[name] = value;
            ***REMOVED***
        ***REMOVED***

        if (typeof this.include === 'string') {
            this.include = [this.include];
        ***REMOVED***

        if (typeof this.exclude === 'string') {
            this.exclude = [this.exclude];
        ***REMOVED***

        if (typeof this.extension === 'string') {
            this.extension = [this.extension];
        ***REMOVED*** else if (this.extension.length === 0) {
            this.extension = false;
        ***REMOVED***

        if (this.include && this.include.length > 0) {
            this.include = prepGlobPatterns([].concat(this.include));
        ***REMOVED*** else {
            this.include = false;
        ***REMOVED***

        if (
            this.excludeNodeModules &&
            !this.exclude.includes('**/node_modules/**')
        ) {
            this.exclude = this.exclude.concat('**/node_modules/**');
        ***REMOVED***

        this.exclude = prepGlobPatterns([].concat(this.exclude));

        this.handleNegation();
    ***REMOVED***

    /* handle the special case of negative globs
     * (!**foo/bar); we create a new this.excludeNegated set
     * of rules, which is applied after excludes and we
     * move excluded include rules into this.excludes.
     */
    handleNegation() {
        const noNeg = e => e.charAt(0) !== '!';
        const onlyNeg = e => e.charAt(0) === '!';
        const stripNeg = e => e.slice(1);

        if (Array.isArray(this.include)) {
            const includeNegated = this.include.filter(onlyNeg).map(stripNeg);
            this.exclude.push(...prepGlobPatterns(includeNegated));
            this.include = this.include.filter(noNeg);
        ***REMOVED***

        this.excludeNegated = this.exclude.filter(onlyNeg).map(stripNeg);
        this.exclude = this.exclude.filter(noNeg);
        this.excludeNegated = prepGlobPatterns(this.excludeNegated);
    ***REMOVED***

    shouldInstrument(filename, relFile) {
        if (
            this.extension &&
            !this.extension.some(ext => filename.endsWith(ext))
        ) {
            return false;
        ***REMOVED***

        let pathToCheck = filename;

        if (this.relativePath) {
            relFile = relFile || path.relative(this.cwd, filename);

            // Don't instrument files that are outside of the current working directory.
            if (isOutsideDir(this.cwd, filename)) {
                return false;
            ***REMOVED***

            pathToCheck = relFile.replace(/^\.[\\/]/, ''); // remove leading './' or '.\'.
        ***REMOVED***

        const dot = { dot: true ***REMOVED***;
        const matches = pattern => minimatch(pathToCheck, pattern, dot);
        return (
            (!this.include || this.include.some(matches)) &&
            (!this.exclude.some(matches) || this.excludeNegated.some(matches))
        );
    ***REMOVED***

    globSync(cwd = this.cwd) {
        const globPatterns = getExtensionPattern(this.extension || []);
        const globOptions = { cwd, nodir: true, dot: true ***REMOVED***;
        /* If we don't have any excludeNegated then we can optimize glob by telling
         * it to not iterate into unwanted directory trees (like node_modules). */
        if (this.excludeNegated.length === 0) {
            globOptions.ignore = this.exclude;
        ***REMOVED***

        return glob
            .sync(globPatterns, globOptions)
            .filter(file => this.shouldInstrument(path.resolve(cwd, file)));
    ***REMOVED***

    async glob(cwd = this.cwd) {
        const globPatterns = getExtensionPattern(this.extension || []);
        const globOptions = { cwd, nodir: true, dot: true ***REMOVED***;
        /* If we don't have any excludeNegated then we can optimize glob by telling
         * it to not iterate into unwanted directory trees (like node_modules). */
        if (this.excludeNegated.length === 0) {
            globOptions.ignore = this.exclude;
        ***REMOVED***

        const list = await glob(globPatterns, globOptions);
        return list.filter(file => this.shouldInstrument(path.resolve(cwd, file)));
    ***REMOVED***
***REMOVED***

function prepGlobPatterns(patterns) {
    return patterns.reduce((result, pattern) => {
        // Allow gitignore style of directory exclusion
        if (!/\/\*\*$/.test(pattern)) {
            result = result.concat(pattern.replace(/\/$/, '') + '/**');
        ***REMOVED***

        // Any rules of the form **/foo.js, should also match foo.js.
        if (/^\*\*\//.test(pattern)) {
            result = result.concat(pattern.replace(/^\*\*\//, ''));
        ***REMOVED***

        return result.concat(pattern);
    ***REMOVED***, []);
***REMOVED***

function getExtensionPattern(extension) {
    switch (extension.length) {
        case 0:
            return '**';
        case 1:
            return `**/*${extension[0]***REMOVED***`;
        default:
            return `**/*{${extension.join()***REMOVED******REMOVED***`;
    ***REMOVED***
***REMOVED***

module.exports = TestExclude;
