// tailwindcss/lib/cli/build/utils.js
// @ts-check
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
***REMOVED***);
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    ***REMOVED***);
***REMOVED***
_export(exports, {
    indentRecursive: function() {
        return indentRecursive;
    ***REMOVED***,
    formatNodes: function() {
        return formatNodes;
    ***REMOVED***,
    readFileWithRetries: function() {
        return readFileWithRetries;
    ***REMOVED***,
    drainStdin: function() {
        return drainStdin;
    ***REMOVED***,
    outputFile: function() {
        return outputFile;
    ***REMOVED***
***REMOVED***);
const _fs = /*#__PURE__*/ _interop_require_default(require("fs"));
const _path = /*#__PURE__*/ _interop_require_default(require("path"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    ***REMOVED***;
***REMOVED***
function indentRecursive(node, indent = 0) {
    node.each && node.each((child, i)=>{
        if (!child.raws.before || !child.raws.before.trim() || child.raws.before.includes("\n")) {
            child.raws.before = `\n${node.type !== "rule" && i > 0 ? "\n" : ""***REMOVED***${"  ".repeat(indent)***REMOVED***`;
        ***REMOVED***
        child.raws.after = `\n${"  ".repeat(indent)***REMOVED***`;
        indentRecursive(child, indent + 1);
    ***REMOVED***);
***REMOVED***
function formatNodes(root) {
    indentRecursive(root);
    if (root.first) {
        root.first.raws.before = "";
    ***REMOVED***
***REMOVED***
async function readFileWithRetries(path, tries = 5) {
    for(let n = 0; n <= tries; n++){
        try {
            return await _fs.default.promises.readFile(path, "utf8");
        ***REMOVED*** catch (err) {
            if (n !== tries) {
                if (err.code === "ENOENT" || err.code === "EBUSY") {
                    await new Promise((resolve)=>setTimeout(resolve, 10));
                    continue;
                ***REMOVED***
            ***REMOVED***
            throw err;
        ***REMOVED***
    ***REMOVED***
***REMOVED***
function drainStdin() {
    return new Promise((resolve, reject)=>{
        let result = "";
        process.stdin.on("data", (chunk)=>{
            result += chunk;
        ***REMOVED***);
        process.stdin.on("end", ()=>resolve(result));
        process.stdin.on("error", (err)=>reject(err));
    ***REMOVED***);
***REMOVED***
async function outputFile(file, newContents) {
    try {
        let currentContents = await _fs.default.promises.readFile(file, "utf8");
        if (currentContents === newContents) {
            return; // Skip writing the file
        ***REMOVED***
    ***REMOVED*** catch  {***REMOVED***
    // Write the file
    await _fs.default.promises.mkdir(_path.default.dirname(file), {
        recursive: true
    ***REMOVED***);
    await _fs.default.promises.writeFile(file, newContents, "utf8");
***REMOVED***
