// workbox-strategies/build/workbox-strategies.dev.js
this.workbox = this.workbox || {***REMOVED***;
this.workbox.strategies = (function (exports, assert_js, logger_js, WorkboxError_js, cacheNames_js, getFriendlyURL_js, cacheMatchIgnoreParams_js, Deferred_js, executeQuotaErrorCallbacks_js, timeout_js) {
    'use strict';

    try {
      self['workbox:strategies:6.5.4'] && _();
    ***REMOVED*** catch (e) {***REMOVED***

    /*
      Copyright 2020 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */

    function toRequest(input) {
      return typeof input === 'string' ? new Request(input) : input;
    ***REMOVED***
    /**
     * A class created every time a Strategy instance instance calls
     * {@link workbox-strategies.Strategy~handle***REMOVED*** or
     * {@link workbox-strategies.Strategy~handleAll***REMOVED*** that wraps all fetch and
     * cache actions around plugin callbacks and keeps track of when the strategy
     * is "done" (i.e. all added `event.waitUntil()` promises have resolved).
     *
     * @memberof workbox-strategies
     */


    class StrategyHandler {
      /**
       * Creates a new instance associated with the passed strategy and event
       * that's handling the request.
       *
       * The constructor also initializes the state that will be passed to each of
       * the plugins handling this request.
       *
       * @param {workbox-strategies.Strategy***REMOVED*** strategy
       * @param {Object***REMOVED*** options
       * @param {Request|string***REMOVED*** options.request A request to run this strategy for.
       * @param {ExtendableEvent***REMOVED*** options.event The event associated with the
       *     request.
       * @param {URL***REMOVED*** [options.url]
       * @param {****REMOVED*** [options.params] The return value from the
       *     {@link workbox-routing~matchCallback***REMOVED*** (if applicable).
       */
      constructor(strategy, options) {
        this._cacheKeys = {***REMOVED***;
        /**
         * The request the strategy is performing (passed to the strategy's
         * `handle()` or `handleAll()` method).
         * @name request
         * @instance
         * @type {Request***REMOVED***
         * @memberof workbox-strategies.StrategyHandler
         */

        /**
         * The event associated with this request.
         * @name event
         * @instance
         * @type {ExtendableEvent***REMOVED***
         * @memberof workbox-strategies.StrategyHandler
         */

        /**
         * A `URL` instance of `request.url` (if passed to the strategy's
         * `handle()` or `handleAll()` method).
         * Note: the `url` param will be present if the strategy was invoked
         * from a workbox `Route` object.
         * @name url
         * @instance
         * @type {URL|undefined***REMOVED***
         * @memberof workbox-strategies.StrategyHandler
         */

        /**
         * A `param` value (if passed to the strategy's
         * `handle()` or `handleAll()` method).
         * Note: the `param` param will be present if the strategy was invoked
         * from a workbox `Route` object and the
         * {@link workbox-routing~matchCallback***REMOVED*** returned
         * a truthy value (it will be that value).
         * @name params
         * @instance
         * @type {*|undefined***REMOVED***
         * @memberof workbox-strategies.StrategyHandler
         */

        {
          assert_js.assert.isInstance(options.event, ExtendableEvent, {
            moduleName: 'workbox-strategies',
            className: 'StrategyHandler',
            funcName: 'constructor',
            paramName: 'options.event'
          ***REMOVED***);
        ***REMOVED***

        Object.assign(this, options);
        this.event = options.event;
        this._strategy = strategy;
        this._handlerDeferred = new Deferred_js.Deferred();
        this._extendLifetimePromises = []; // Copy the plugins list (since it's mutable on the strategy),
        // so any mutations don't affect this handler instance.

        this._plugins = [...strategy.plugins];
        this._pluginStateMap = new Map();

        for (const plugin of this._plugins) {
          this._pluginStateMap.set(plugin, {***REMOVED***);
        ***REMOVED***

        this.event.waitUntil(this._handlerDeferred.promise);
      ***REMOVED***
      /**
       * Fetches a given request (and invokes any applicable plugin callback
       * methods) using the `fetchOptions` (for non-navigation requests) and
       * `plugins` defined on the `Strategy` object.
       *
       * The following plugin lifecycle methods are invoked when using this method:
       * - `requestWillFetch()`
       * - `fetchDidSucceed()`
       * - `fetchDidFail()`
       *
       * @param {Request|string***REMOVED*** input The URL or request to fetch.
       * @return {Promise<Response>***REMOVED***
       */


      async fetch(input) {
        const {
          event
        ***REMOVED*** = this;
        let request = toRequest(input);

        if (request.mode === 'navigate' && event instanceof FetchEvent && event.preloadResponse) {
          const possiblePreloadResponse = await event.preloadResponse;

          if (possiblePreloadResponse) {
            {
              logger_js.logger.log(`Using a preloaded navigation response for ` + `'${getFriendlyURL_js.getFriendlyURL(request.url)***REMOVED***'`);
            ***REMOVED***

            return possiblePreloadResponse;
          ***REMOVED***
        ***REMOVED*** // If there is a fetchDidFail plugin, we need to save a clone of the
        // original request before it's either modified by a requestWillFetch
        // plugin or before the original request's body is consumed via fetch().


        const originalRequest = this.hasCallback('fetchDidFail') ? request.clone() : null;

        try {
          for (const cb of this.iterateCallbacks('requestWillFetch')) {
            request = await cb({
              request: request.clone(),
              event
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED*** catch (err) {
          if (err instanceof Error) {
            throw new WorkboxError_js.WorkboxError('plugin-error-request-will-fetch', {
              thrownErrorMessage: err.message
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED*** // The request can be altered by plugins with `requestWillFetch` making
        // the original request (most likely from a `fetch` event) different
        // from the Request we make. Pass both to `fetchDidFail` to aid debugging.


        const pluginFilteredRequest = request.clone();

        try {
          let fetchResponse; // See https://github.com/GoogleChrome/workbox/issues/1796

          fetchResponse = await fetch(request, request.mode === 'navigate' ? undefined : this._strategy.fetchOptions);

          if ("dev" !== 'production') {
            logger_js.logger.debug(`Network request for ` + `'${getFriendlyURL_js.getFriendlyURL(request.url)***REMOVED***' returned a response with ` + `status '${fetchResponse.status***REMOVED***'.`);
          ***REMOVED***

          for (const callback of this.iterateCallbacks('fetchDidSucceed')) {
            fetchResponse = await callback({
              event,
              request: pluginFilteredRequest,
              response: fetchResponse
            ***REMOVED***);
          ***REMOVED***

          return fetchResponse;
        ***REMOVED*** catch (error) {
          {
            logger_js.logger.log(`Network request for ` + `'${getFriendlyURL_js.getFriendlyURL(request.url)***REMOVED***' threw an error.`, error);
          ***REMOVED*** // `originalRequest` will only exist if a `fetchDidFail` callback
          // is being used (see above).


          if (originalRequest) {
            await this.runCallbacks('fetchDidFail', {
              error: error,
              event,
              originalRequest: originalRequest.clone(),
              request: pluginFilteredRequest.clone()
            ***REMOVED***);
          ***REMOVED***

          throw error;
        ***REMOVED***
      ***REMOVED***
      /**
       * Calls `this.fetch()` and (in the background) runs `this.cachePut()` on
       * the response generated by `this.fetch()`.
       *
       * The call to `this.cachePut()` automatically invokes `this.waitUntil()`,
       * so you do not have to manually call `waitUntil()` on the event.
       *
       * @param {Request|string***REMOVED*** input The request or URL to fetch and cache.
       * @return {Promise<Response>***REMOVED***
       */


      async fetchAndCachePut(input) {
        const response = await this.fetch(input);
        const responseClone = response.clone();
        void this.waitUntil(this.cachePut(input, responseClone));
        return response;
      ***REMOVED***
      /**
       * Matches a request from the cache (and invokes any applicable plugin
       * callback methods) using the `cacheName`, `matchOptions`, and `plugins`
       * defined on the strategy object.
       *
       * The following plugin lifecycle methods are invoked when using this method:
       * - cacheKeyWillByUsed()
       * - cachedResponseWillByUsed()
       *
       * @param {Request|string***REMOVED*** key The Request or URL to use as the cache key.
       * @return {Promise<Response|undefined>***REMOVED*** A matching response, if found.
       */


      async cacheMatch(key) {
        const request = toRequest(key);
        let cachedResponse;
        const {
          cacheName,
          matchOptions
        ***REMOVED*** = this._strategy;
        const effectiveRequest = await this.getCacheKey(request, 'read');
        const multiMatchOptions = Object.assign(Object.assign({***REMOVED***, matchOptions), {
          cacheName
        ***REMOVED***);
        cachedResponse = await caches.match(effectiveRequest, multiMatchOptions);

        {
          if (cachedResponse) {
            logger_js.logger.debug(`Found a cached response in '${cacheName***REMOVED***'.`);
          ***REMOVED*** else {
            logger_js.logger.debug(`No cached response found in '${cacheName***REMOVED***'.`);
          ***REMOVED***
        ***REMOVED***

        for (const callback of this.iterateCallbacks('cachedResponseWillBeUsed')) {
          cachedResponse = (await callback({
            cacheName,
            matchOptions,
            cachedResponse,
            request: effectiveRequest,
            event: this.event
          ***REMOVED***)) || undefined;
        ***REMOVED***

        return cachedResponse;
      ***REMOVED***
      /**
       * Puts a request/response pair in the cache (and invokes any applicable
       * plugin callback methods) using the `cacheName` and `plugins` defined on
       * the strategy object.
       *
       * The following plugin lifecycle methods are invoked when using this method:
       * - cacheKeyWillByUsed()
       * - cacheWillUpdate()
       * - cacheDidUpdate()
       *
       * @param {Request|string***REMOVED*** key The request or URL to use as the cache key.
       * @param {Response***REMOVED*** response The response to cache.
       * @return {Promise<boolean>***REMOVED*** `false` if a cacheWillUpdate caused the response
       * not be cached, and `true` otherwise.
       */


      async cachePut(key, response) {
        const request = toRequest(key); // Run in the next task to avoid blocking other cache reads.
        // https://github.com/w3c/ServiceWorker/issues/1397

        await timeout_js.timeout(0);
        const effectiveRequest = await this.getCacheKey(request, 'write');

        {
          if (effectiveRequest.method && effectiveRequest.method !== 'GET') {
            throw new WorkboxError_js.WorkboxError('attempt-to-cache-non-get-request', {
              url: getFriendlyURL_js.getFriendlyURL(effectiveRequest.url),
              method: effectiveRequest.method
            ***REMOVED***);
          ***REMOVED*** // See https://github.com/GoogleChrome/workbox/issues/2818


          const vary = response.headers.get('Vary');

          if (vary) {
            logger_js.logger.debug(`The response for ${getFriendlyURL_js.getFriendlyURL(effectiveRequest.url)***REMOVED*** ` + `has a 'Vary: ${vary***REMOVED***' header. ` + `Consider setting the {ignoreVary: true***REMOVED*** option on your strategy ` + `to ensure cache matching and deletion works as expected.`);
          ***REMOVED***
        ***REMOVED***

        if (!response) {
          {
            logger_js.logger.error(`Cannot cache non-existent response for ` + `'${getFriendlyURL_js.getFriendlyURL(effectiveRequest.url)***REMOVED***'.`);
          ***REMOVED***

          throw new WorkboxError_js.WorkboxError('cache-put-with-no-response', {
            url: getFriendlyURL_js.getFriendlyURL(effectiveRequest.url)
          ***REMOVED***);
        ***REMOVED***

        const responseToCache = await this._ensureResponseSafeToCache(response);

        if (!responseToCache) {
          {
            logger_js.logger.debug(`Response '${getFriendlyURL_js.getFriendlyURL(effectiveRequest.url)***REMOVED***' ` + `will not be cached.`, responseToCache);
          ***REMOVED***

          return false;
        ***REMOVED***

        const {
          cacheName,
          matchOptions
        ***REMOVED*** = this._strategy;
        const cache = await self.caches.open(cacheName);
        const hasCacheUpdateCallback = this.hasCallback('cacheDidUpdate');
        const oldResponse = hasCacheUpdateCallback ? await cacheMatchIgnoreParams_js.cacheMatchIgnoreParams( // TODO(philipwalton): the `__WB_REVISION__` param is a precaching
        // feature. Consider into ways to only add this behavior if using
        // precaching.
        cache, effectiveRequest.clone(), ['__WB_REVISION__'], matchOptions) : null;

        {
          logger_js.logger.debug(`Updating the '${cacheName***REMOVED***' cache with a new Response ` + `for ${getFriendlyURL_js.getFriendlyURL(effectiveRequest.url)***REMOVED***.`);
        ***REMOVED***

        try {
          await cache.put(effectiveRequest, hasCacheUpdateCallback ? responseToCache.clone() : responseToCache);
        ***REMOVED*** catch (error) {
          if (error instanceof Error) {
            // See https://developer.mozilla.org/en-US/docs/Web/API/DOMException#exception-QuotaExceededError
            if (error.name === 'QuotaExceededError') {
              await executeQuotaErrorCallbacks_js.executeQuotaErrorCallbacks();
            ***REMOVED***

            throw error;
          ***REMOVED***
        ***REMOVED***

        for (const callback of this.iterateCallbacks('cacheDidUpdate')) {
          await callback({
            cacheName,
            oldResponse,
            newResponse: responseToCache.clone(),
            request: effectiveRequest,
            event: this.event
          ***REMOVED***);
        ***REMOVED***

        return true;
      ***REMOVED***
      /**
       * Checks the list of plugins for the `cacheKeyWillBeUsed` callback, and
       * executes any of those callbacks found in sequence. The final `Request`
       * object returned by the last plugin is treated as the cache key for cache
       * reads and/or writes. If no `cacheKeyWillBeUsed` plugin callbacks have
       * been registered, the passed request is returned unmodified
       *
       * @param {Request***REMOVED*** request
       * @param {string***REMOVED*** mode
       * @return {Promise<Request>***REMOVED***
       */


      async getCacheKey(request, mode) {
        const key = `${request.url***REMOVED*** | ${mode***REMOVED***`;

        if (!this._cacheKeys[key]) {
          let effectiveRequest = request;

          for (const callback of this.iterateCallbacks('cacheKeyWillBeUsed')) {
            effectiveRequest = toRequest(await callback({
              mode,
              request: effectiveRequest,
              event: this.event,
              // params has a type any can't change right now.
              params: this.params // eslint-disable-line

            ***REMOVED***));
          ***REMOVED***

          this._cacheKeys[key] = effectiveRequest;
        ***REMOVED***

        return this._cacheKeys[key];
      ***REMOVED***
      /**
       * Returns true if the strategy has at least one plugin with the given
       * callback.
       *
       * @param {string***REMOVED*** name The name of the callback to check for.
       * @return {boolean***REMOVED***
       */


      hasCallback(name) {
        for (const plugin of this._strategy.plugins) {
          if (name in plugin) {
            return true;
          ***REMOVED***
        ***REMOVED***

        return false;
      ***REMOVED***
      /**
       * Runs all plugin callbacks matching the given name, in order, passing the
       * given param object (merged ith the current plugin state) as the only
       * argument.
       *
       * Note: since this method runs all plugins, it's not suitable for cases
       * where the return value of a callback needs to be applied prior to calling
       * the next callback. See
       * {@link workbox-strategies.StrategyHandler#iterateCallbacks***REMOVED***
       * below for how to handle that case.
       *
       * @param {string***REMOVED*** name The name of the callback to run within each plugin.
       * @param {Object***REMOVED*** param The object to pass as the first (and only) param
       *     when executing each callback. This object will be merged with the
       *     current plugin state prior to callback execution.
       */


      async runCallbacks(name, param) {
        for (const callback of this.iterateCallbacks(name)) {
          // TODO(philipwalton): not sure why `any` is needed. It seems like
          // this should work with `as WorkboxPluginCallbackParam[C]`.
          await callback(param);
        ***REMOVED***
      ***REMOVED***
      /**
       * Accepts a callback and returns an iterable of matching plugin callbacks,
       * where each callback is wrapped with the current handler state (i.e. when
       * you call each callback, whatever object parameter you pass it will
       * be merged with the plugin's current state).
       *
       * @param {string***REMOVED*** name The name fo the callback to run
       * @return {Array<Function>***REMOVED***
       */


      *iterateCallbacks(name) {
        for (const plugin of this._strategy.plugins) {
          if (typeof plugin[name] === 'function') {
            const state = this._pluginStateMap.get(plugin);

            const statefulCallback = param => {
              const statefulParam = Object.assign(Object.assign({***REMOVED***, param), {
                state
              ***REMOVED***); // TODO(philipwalton): not sure why `any` is needed. It seems like
              // this should work with `as WorkboxPluginCallbackParam[C]`.

              return plugin[name](statefulParam);
            ***REMOVED***;

            yield statefulCallback;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
      /**
       * Adds a promise to the
       * [extend lifetime promises]{@link https://w3c.github.io/ServiceWorker/#extendableevent-extend-lifetime-promises***REMOVED***
       * of the event event associated with the request being handled (usually a
       * `FetchEvent`).
       *
       * Note: you can await
       * {@link workbox-strategies.StrategyHandler~doneWaiting***REMOVED***
       * to know when all added promises have settled.
       *
       * @param {Promise***REMOVED*** promise A promise to add to the extend lifetime promises
       *     of the event that triggered the request.
       */


      waitUntil(promise) {
        this._extendLifetimePromises.push(promise);

        return promise;
      ***REMOVED***
      /**
       * Returns a promise that resolves once all promises passed to
       * {@link workbox-strategies.StrategyHandler~waitUntil***REMOVED***
       * have settled.
       *
       * Note: any work done after `doneWaiting()` settles should be manually
       * passed to an event's `waitUntil()` method (not this handler's
       * `waitUntil()` method), otherwise the service worker thread my be killed
       * prior to your work completing.
       */


      async doneWaiting() {
        let promise;

        while (promise = this._extendLifetimePromises.shift()) {
          await promise;
        ***REMOVED***
      ***REMOVED***
      /**
       * Stops running the strategy and immediately resolves any pending
       * `waitUntil()` promises.
       */


      destroy() {
        this._handlerDeferred.resolve(null);
      ***REMOVED***
      /**
       * This method will call cacheWillUpdate on the available plugins (or use
       * status === 200) to determine if the Response is safe and valid to cache.
       *
       * @param {Request***REMOVED*** options.request
       * @param {Response***REMOVED*** options.response
       * @return {Promise<Response|undefined>***REMOVED***
       *
       * @private
       */


      async _ensureResponseSafeToCache(response) {
        let responseToCache = response;
        let pluginsUsed = false;

        for (const callback of this.iterateCallbacks('cacheWillUpdate')) {
          responseToCache = (await callback({
            request: this.request,
            response: responseToCache,
            event: this.event
          ***REMOVED***)) || undefined;
          pluginsUsed = true;

          if (!responseToCache) {
            break;
          ***REMOVED***
        ***REMOVED***

        if (!pluginsUsed) {
          if (responseToCache && responseToCache.status !== 200) {
            responseToCache = undefined;
          ***REMOVED***

          {
            if (responseToCache) {
              if (responseToCache.status !== 200) {
                if (responseToCache.status === 0) {
                  logger_js.logger.warn(`The response for '${this.request.url***REMOVED***' ` + `is an opaque response. The caching strategy that you're ` + `using will not cache opaque responses by default.`);
                ***REMOVED*** else {
                  logger_js.logger.debug(`The response for '${this.request.url***REMOVED***' ` + `returned a status code of '${response.status***REMOVED***' and won't ` + `be cached as a result.`);
                ***REMOVED***
              ***REMOVED***
            ***REMOVED***
          ***REMOVED***
        ***REMOVED***

        return responseToCache;
      ***REMOVED***

    ***REMOVED***

    /*
      Copyright 2020 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /**
     * An abstract base class that all other strategy classes must extend from:
     *
     * @memberof workbox-strategies
     */

    class Strategy {
      /**
       * Creates a new instance of the strategy and sets all documented option
       * properties as public instance properties.
       *
       * Note: if a custom strategy class extends the base Strategy class and does
       * not need more than these properties, it does not need to define its own
       * constructor.
       *
       * @param {Object***REMOVED*** [options]
       * @param {string***REMOVED*** [options.cacheName] Cache name to store and retrieve
       * requests. Defaults to the cache names provided by
       * {@link workbox-core.cacheNames***REMOVED***.
       * @param {Array<Object>***REMOVED*** [options.plugins] [Plugins]{@link https://developers.google.com/web/tools/workbox/guides/using-plugins***REMOVED***
       * to use in conjunction with this caching strategy.
       * @param {Object***REMOVED*** [options.fetchOptions] Values passed along to the
       * [`init`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch#Parameters)
       * of [non-navigation](https://github.com/GoogleChrome/workbox/issues/1796)
       * `fetch()` requests made by this strategy.
       * @param {Object***REMOVED*** [options.matchOptions] The
       * [`CacheQueryOptions`]{@link https://w3c.github.io/ServiceWorker/#dictdef-cachequeryoptions***REMOVED***
       * for any `cache.match()` or `cache.put()` calls made by this strategy.
       */
      constructor(options = {***REMOVED***) {
        /**
         * Cache name to store and retrieve
         * requests. Defaults to the cache names provided by
         * {@link workbox-core.cacheNames***REMOVED***.
         *
         * @type {string***REMOVED***
         */
        this.cacheName = cacheNames_js.cacheNames.getRuntimeName(options.cacheName);
        /**
         * The list
         * [Plugins]{@link https://developers.google.com/web/tools/workbox/guides/using-plugins***REMOVED***
         * used by this strategy.
         *
         * @type {Array<Object>***REMOVED***
         */

        this.plugins = options.plugins || [];
        /**
         * Values passed along to the
         * [`init`]{@link https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch#Parameters***REMOVED***
         * of all fetch() requests made by this strategy.
         *
         * @type {Object***REMOVED***
         */

        this.fetchOptions = options.fetchOptions;
        /**
         * The
         * [`CacheQueryOptions`]{@link https://w3c.github.io/ServiceWorker/#dictdef-cachequeryoptions***REMOVED***
         * for any `cache.match()` or `cache.put()` calls made by this strategy.
         *
         * @type {Object***REMOVED***
         */

        this.matchOptions = options.matchOptions;
      ***REMOVED***
      /**
       * Perform a request strategy and returns a `Promise` that will resolve with
       * a `Response`, invoking all relevant plugin callbacks.
       *
       * When a strategy instance is registered with a Workbox
       * {@link workbox-routing.Route***REMOVED***, this method is automatically
       * called when the route matches.
       *
       * Alternatively, this method can be used in a standalone `FetchEvent`
       * listener by passing it to `event.respondWith()`.
       *
       * @param {FetchEvent|Object***REMOVED*** options A `FetchEvent` or an object with the
       *     properties listed below.
       * @param {Request|string***REMOVED*** options.request A request to run this strategy for.
       * @param {ExtendableEvent***REMOVED*** options.event The event associated with the
       *     request.
       * @param {URL***REMOVED*** [options.url]
       * @param {****REMOVED*** [options.params]
       */


      handle(options) {
        const [responseDone] = this.handleAll(options);
        return responseDone;
      ***REMOVED***
      /**
       * Similar to {@link workbox-strategies.Strategy~handle***REMOVED***, but
       * instead of just returning a `Promise` that resolves to a `Response` it
       * it will return an tuple of `[response, done]` promises, where the former
       * (`response`) is equivalent to what `handle()` returns, and the latter is a
       * Promise that will resolve once any promises that were added to
       * `event.waitUntil()` as part of performing the strategy have completed.
       *
       * You can await the `done` promise to ensure any extra work performed by
       * the strategy (usually caching responses) completes successfully.
       *
       * @param {FetchEvent|Object***REMOVED*** options A `FetchEvent` or an object with the
       *     properties listed below.
       * @param {Request|string***REMOVED*** options.request A request to run this strategy for.
       * @param {ExtendableEvent***REMOVED*** options.event The event associated with the
       *     request.
       * @param {URL***REMOVED*** [options.url]
       * @param {****REMOVED*** [options.params]
       * @return {Array<Promise>***REMOVED*** A tuple of [response, done]
       *     promises that can be used to determine when the response resolves as
       *     well as when the handler has completed all its work.
       */


      handleAll(options) {
        // Allow for flexible options to be passed.
        if (options instanceof FetchEvent) {
          options = {
            event: options,
            request: options.request
          ***REMOVED***;
        ***REMOVED***

        const event = options.event;
        const request = typeof options.request === 'string' ? new Request(options.request) : options.request;
        const params = 'params' in options ? options.params : undefined;
        const handler = new StrategyHandler(this, {
          event,
          request,
          params
        ***REMOVED***);

        const responseDone = this._getResponse(handler, request, event);

        const handlerDone = this._awaitComplete(responseDone, handler, request, event); // Return an array of promises, suitable for use with Promise.all().


        return [responseDone, handlerDone];
      ***REMOVED***

      async _getResponse(handler, request, event) {
        await handler.runCallbacks('handlerWillStart', {
          event,
          request
        ***REMOVED***);
        let response = undefined;

        try {
          response = await this._handle(request, handler); // The "official" Strategy subclasses all throw this error automatically,
          // but in case a third-party Strategy doesn't, ensure that we have a
          // consistent failure when there's no response or an error response.

          if (!response || response.type === 'error') {
            throw new WorkboxError_js.WorkboxError('no-response', {
              url: request.url
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED*** catch (error) {
          if (error instanceof Error) {
            for (const callback of handler.iterateCallbacks('handlerDidError')) {
              response = await callback({
                error,
                event,
                request
              ***REMOVED***);

              if (response) {
                break;
              ***REMOVED***
            ***REMOVED***
          ***REMOVED***

          if (!response) {
            throw error;
          ***REMOVED*** else {
            logger_js.logger.log(`While responding to '${getFriendlyURL_js.getFriendlyURL(request.url)***REMOVED***', ` + `an ${error instanceof Error ? error.toString() : ''***REMOVED*** error occurred. Using a fallback response provided by ` + `a handlerDidError plugin.`);
          ***REMOVED***
        ***REMOVED***

        for (const callback of handler.iterateCallbacks('handlerWillRespond')) {
          response = await callback({
            event,
            request,
            response
          ***REMOVED***);
        ***REMOVED***

        return response;
      ***REMOVED***

      async _awaitComplete(responseDone, handler, request, event) {
        let response;
        let error;

        try {
          response = await responseDone;
        ***REMOVED*** catch (error) {// Ignore errors, as response errors should be caught via the `response`
          // promise above. The `done` promise will only throw for errors in
          // promises passed to `handler.waitUntil()`.
        ***REMOVED***

        try {
          await handler.runCallbacks('handlerDidRespond', {
            event,
            request,
            response
          ***REMOVED***);
          await handler.doneWaiting();
        ***REMOVED*** catch (waitUntilError) {
          if (waitUntilError instanceof Error) {
            error = waitUntilError;
          ***REMOVED***
        ***REMOVED***

        await handler.runCallbacks('handlerDidComplete', {
          event,
          request,
          response,
          error: error
        ***REMOVED***);
        handler.destroy();

        if (error) {
          throw error;
        ***REMOVED***
      ***REMOVED***

    ***REMOVED***
    /**
     * Classes extending the `Strategy` based class should implement this method,
     * and leverage the {@link workbox-strategies.StrategyHandler***REMOVED***
     * arg to perform all fetching and cache logic, which will ensure all relevant
     * cache, cache options, fetch options and plugins are used (per the current
     * strategy instance).
     *
     * @name _handle
     * @instance
     * @abstract
     * @function
     * @param {Request***REMOVED*** request
     * @param {workbox-strategies.StrategyHandler***REMOVED*** handler
     * @return {Promise<Response>***REMOVED***
     *
     * @memberof workbox-strategies.Strategy
     */

    /*
      Copyright 2018 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    const messages = {
      strategyStart: (strategyName, request) => `Using ${strategyName***REMOVED*** to respond to '${getFriendlyURL_js.getFriendlyURL(request.url)***REMOVED***'`,
      printFinalResponse: response => {
        if (response) {
          logger_js.logger.groupCollapsed(`View the final response here.`);
          logger_js.logger.log(response || '[No response returned]');
          logger_js.logger.groupEnd();
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***;

    /*
      Copyright 2018 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /**
     * An implementation of a [cache-first](https://developer.chrome.com/docs/workbox/caching-strategies-overview/#cache-first-falling-back-to-network)
     * request strategy.
     *
     * A cache first strategy is useful for assets that have been revisioned,
     * such as URLs like `/styles/example.a8f5f1.css`, since they
     * can be cached for long periods of time.
     *
     * If the network request fails, and there is no cache match, this will throw
     * a `WorkboxError` exception.
     *
     * @extends workbox-strategies.Strategy
     * @memberof workbox-strategies
     */

    class CacheFirst extends Strategy {
      /**
       * @private
       * @param {Request|string***REMOVED*** request A request to run this strategy for.
       * @param {workbox-strategies.StrategyHandler***REMOVED*** handler The event that
       *     triggered the request.
       * @return {Promise<Response>***REMOVED***
       */
      async _handle(request, handler) {
        const logs = [];

        {
          assert_js.assert.isInstance(request, Request, {
            moduleName: 'workbox-strategies',
            className: this.constructor.name,
            funcName: 'makeRequest',
            paramName: 'request'
          ***REMOVED***);
        ***REMOVED***

        let response = await handler.cacheMatch(request);
        let error = undefined;

        if (!response) {
          {
            logs.push(`No response found in the '${this.cacheName***REMOVED***' cache. ` + `Will respond with a network request.`);
          ***REMOVED***

          try {
            response = await handler.fetchAndCachePut(request);
          ***REMOVED*** catch (err) {
            if (err instanceof Error) {
              error = err;
            ***REMOVED***
          ***REMOVED***

          {
            if (response) {
              logs.push(`Got response from network.`);
            ***REMOVED*** else {
              logs.push(`Unable to get a response from the network.`);
            ***REMOVED***
          ***REMOVED***
        ***REMOVED*** else {
          {
            logs.push(`Found a cached response in the '${this.cacheName***REMOVED***' cache.`);
          ***REMOVED***
        ***REMOVED***

        {
          logger_js.logger.groupCollapsed(messages.strategyStart(this.constructor.name, request));

          for (const log of logs) {
            logger_js.logger.log(log);
          ***REMOVED***

          messages.printFinalResponse(response);
          logger_js.logger.groupEnd();
        ***REMOVED***

        if (!response) {
          throw new WorkboxError_js.WorkboxError('no-response', {
            url: request.url,
            error
          ***REMOVED***);
        ***REMOVED***

        return response;
      ***REMOVED***

    ***REMOVED***

    /*
      Copyright 2018 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /**
     * An implementation of a [cache-only](https://developer.chrome.com/docs/workbox/caching-strategies-overview/#cache-only)
     * request strategy.
     *
     * This class is useful if you want to take advantage of any
     * [Workbox plugins](https://developer.chrome.com/docs/workbox/using-plugins/).
     *
     * If there is no cache match, this will throw a `WorkboxError` exception.
     *
     * @extends workbox-strategies.Strategy
     * @memberof workbox-strategies
     */

    class CacheOnly extends Strategy {
      /**
       * @private
       * @param {Request|string***REMOVED*** request A request to run this strategy for.
       * @param {workbox-strategies.StrategyHandler***REMOVED*** handler The event that
       *     triggered the request.
       * @return {Promise<Response>***REMOVED***
       */
      async _handle(request, handler) {
        {
          assert_js.assert.isInstance(request, Request, {
            moduleName: 'workbox-strategies',
            className: this.constructor.name,
            funcName: 'makeRequest',
            paramName: 'request'
          ***REMOVED***);
        ***REMOVED***

        const response = await handler.cacheMatch(request);

        {
          logger_js.logger.groupCollapsed(messages.strategyStart(this.constructor.name, request));

          if (response) {
            logger_js.logger.log(`Found a cached response in the '${this.cacheName***REMOVED***' ` + `cache.`);
            messages.printFinalResponse(response);
          ***REMOVED*** else {
            logger_js.logger.log(`No response found in the '${this.cacheName***REMOVED***' cache.`);
          ***REMOVED***

          logger_js.logger.groupEnd();
        ***REMOVED***

        if (!response) {
          throw new WorkboxError_js.WorkboxError('no-response', {
            url: request.url
          ***REMOVED***);
        ***REMOVED***

        return response;
      ***REMOVED***

    ***REMOVED***

    /*
      Copyright 2018 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    const cacheOkAndOpaquePlugin = {
      /**
       * Returns a valid response (to allow caching) if the status is 200 (OK) or
       * 0 (opaque).
       *
       * @param {Object***REMOVED*** options
       * @param {Response***REMOVED*** options.response
       * @return {Response|null***REMOVED***
       *
       * @private
       */
      cacheWillUpdate: async ({
        response
      ***REMOVED***) => {
        if (response.status === 200 || response.status === 0) {
          return response;
        ***REMOVED***

        return null;
      ***REMOVED***
    ***REMOVED***;

    /*
      Copyright 2018 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /**
     * An implementation of a
     * [network first](https://developer.chrome.com/docs/workbox/caching-strategies-overview/#network-first-falling-back-to-cache)
     * request strategy.
     *
     * By default, this strategy will cache responses with a 200 status code as
     * well as [opaque responses](https://developer.chrome.com/docs/workbox/caching-resources-during-runtime/#opaque-responses).
     * Opaque responses are are cross-origin requests where the response doesn't
     * support [CORS](https://enable-cors.org/).
     *
     * If the network request fails, and there is no cache match, this will throw
     * a `WorkboxError` exception.
     *
     * @extends workbox-strategies.Strategy
     * @memberof workbox-strategies
     */

    class NetworkFirst extends Strategy {
      /**
       * @param {Object***REMOVED*** [options]
       * @param {string***REMOVED*** [options.cacheName] Cache name to store and retrieve
       * requests. Defaults to cache names provided by
       * {@link workbox-core.cacheNames***REMOVED***.
       * @param {Array<Object>***REMOVED*** [options.plugins] [Plugins]{@link https://developers.google.com/web/tools/workbox/guides/using-plugins***REMOVED***
       * to use in conjunction with this caching strategy.
       * @param {Object***REMOVED*** [options.fetchOptions] Values passed along to the
       * [`init`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch#Parameters)
       * of [non-navigation](https://github.com/GoogleChrome/workbox/issues/1796)
       * `fetch()` requests made by this strategy.
       * @param {Object***REMOVED*** [options.matchOptions] [`CacheQueryOptions`](https://w3c.github.io/ServiceWorker/#dictdef-cachequeryoptions)
       * @param {number***REMOVED*** [options.networkTimeoutSeconds] If set, any network requests
       * that fail to respond within the timeout will fallback to the cache.
       *
       * This option can be used to combat
       * "[lie-fi]{@link https://developers.google.com/web/fundamentals/performance/poor-connectivity/#lie-fi***REMOVED***"
       * scenarios.
       */
      constructor(options = {***REMOVED***) {
        super(options); // If this instance contains no plugins with a 'cacheWillUpdate' callback,
        // prepend the `cacheOkAndOpaquePlugin` plugin to the plugins list.

        if (!this.plugins.some(p => 'cacheWillUpdate' in p)) {
          this.plugins.unshift(cacheOkAndOpaquePlugin);
        ***REMOVED***

        this._networkTimeoutSeconds = options.networkTimeoutSeconds || 0;

        {
          if (this._networkTimeoutSeconds) {
            assert_js.assert.isType(this._networkTimeoutSeconds, 'number', {
              moduleName: 'workbox-strategies',
              className: this.constructor.name,
              funcName: 'constructor',
              paramName: 'networkTimeoutSeconds'
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
      /**
       * @private
       * @param {Request|string***REMOVED*** request A request to run this strategy for.
       * @param {workbox-strategies.StrategyHandler***REMOVED*** handler The event that
       *     triggered the request.
       * @return {Promise<Response>***REMOVED***
       */


      async _handle(request, handler) {
        const logs = [];

        {
          assert_js.assert.isInstance(request, Request, {
            moduleName: 'workbox-strategies',
            className: this.constructor.name,
            funcName: 'handle',
            paramName: 'makeRequest'
          ***REMOVED***);
        ***REMOVED***

        const promises = [];
        let timeoutId;

        if (this._networkTimeoutSeconds) {
          const {
            id,
            promise
          ***REMOVED*** = this._getTimeoutPromise({
            request,
            logs,
            handler
          ***REMOVED***);

          timeoutId = id;
          promises.push(promise);
        ***REMOVED***

        const networkPromise = this._getNetworkPromise({
          timeoutId,
          request,
          logs,
          handler
        ***REMOVED***);

        promises.push(networkPromise);
        const response = await handler.waitUntil((async () => {
          // Promise.race() will resolve as soon as the first promise resolves.
          return (await handler.waitUntil(Promise.race(promises))) || ( // If Promise.race() resolved with null, it might be due to a network
          // timeout + a cache miss. If that were to happen, we'd rather wait until
          // the networkPromise resolves instead of returning null.
          // Note that it's fine to await an already-resolved promise, so we don't
          // have to check to see if it's still "in flight".
          await networkPromise);
        ***REMOVED***)());

        {
          logger_js.logger.groupCollapsed(messages.strategyStart(this.constructor.name, request));

          for (const log of logs) {
            logger_js.logger.log(log);
          ***REMOVED***

          messages.printFinalResponse(response);
          logger_js.logger.groupEnd();
        ***REMOVED***

        if (!response) {
          throw new WorkboxError_js.WorkboxError('no-response', {
            url: request.url
          ***REMOVED***);
        ***REMOVED***

        return response;
      ***REMOVED***
      /**
       * @param {Object***REMOVED*** options
       * @param {Request***REMOVED*** options.request
       * @param {Array***REMOVED*** options.logs A reference to the logs array
       * @param {Event***REMOVED*** options.event
       * @return {Promise<Response>***REMOVED***
       *
       * @private
       */


      _getTimeoutPromise({
        request,
        logs,
        handler
      ***REMOVED***) {
        let timeoutId;
        const timeoutPromise = new Promise(resolve => {
          const onNetworkTimeout = async () => {
            {
              logs.push(`Timing out the network response at ` + `${this._networkTimeoutSeconds***REMOVED*** seconds.`);
            ***REMOVED***

            resolve(await handler.cacheMatch(request));
          ***REMOVED***;

          timeoutId = setTimeout(onNetworkTimeout, this._networkTimeoutSeconds * 1000);
        ***REMOVED***);
        return {
          promise: timeoutPromise,
          id: timeoutId
        ***REMOVED***;
      ***REMOVED***
      /**
       * @param {Object***REMOVED*** options
       * @param {number|undefined***REMOVED*** options.timeoutId
       * @param {Request***REMOVED*** options.request
       * @param {Array***REMOVED*** options.logs A reference to the logs Array.
       * @param {Event***REMOVED*** options.event
       * @return {Promise<Response>***REMOVED***
       *
       * @private
       */


      async _getNetworkPromise({
        timeoutId,
        request,
        logs,
        handler
      ***REMOVED***) {
        let error;
        let response;

        try {
          response = await handler.fetchAndCachePut(request);
        ***REMOVED*** catch (fetchError) {
          if (fetchError instanceof Error) {
            error = fetchError;
          ***REMOVED***
        ***REMOVED***

        if (timeoutId) {
          clearTimeout(timeoutId);
        ***REMOVED***

        {
          if (response) {
            logs.push(`Got response from network.`);
          ***REMOVED*** else {
            logs.push(`Unable to get a response from the network. Will respond ` + `with a cached response.`);
          ***REMOVED***
        ***REMOVED***

        if (error || !response) {
          response = await handler.cacheMatch(request);

          {
            if (response) {
              logs.push(`Found a cached response in the '${this.cacheName***REMOVED***'` + ` cache.`);
            ***REMOVED*** else {
              logs.push(`No response found in the '${this.cacheName***REMOVED***' cache.`);
            ***REMOVED***
          ***REMOVED***
        ***REMOVED***

        return response;
      ***REMOVED***

    ***REMOVED***

    /*
      Copyright 2018 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /**
     * An implementation of a
     * [network-only](https://developer.chrome.com/docs/workbox/caching-strategies-overview/#network-only)
     * request strategy.
     *
     * This class is useful if you want to take advantage of any
     * [Workbox plugins](https://developer.chrome.com/docs/workbox/using-plugins/).
     *
     * If the network request fails, this will throw a `WorkboxError` exception.
     *
     * @extends workbox-strategies.Strategy
     * @memberof workbox-strategies
     */

    class NetworkOnly extends Strategy {
      /**
       * @param {Object***REMOVED*** [options]
       * @param {Array<Object>***REMOVED*** [options.plugins] [Plugins]{@link https://developers.google.com/web/tools/workbox/guides/using-plugins***REMOVED***
       * to use in conjunction with this caching strategy.
       * @param {Object***REMOVED*** [options.fetchOptions] Values passed along to the
       * [`init`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch#Parameters)
       * of [non-navigation](https://github.com/GoogleChrome/workbox/issues/1796)
       * `fetch()` requests made by this strategy.
       * @param {number***REMOVED*** [options.networkTimeoutSeconds] If set, any network requests
       * that fail to respond within the timeout will result in a network error.
       */
      constructor(options = {***REMOVED***) {
        super(options);
        this._networkTimeoutSeconds = options.networkTimeoutSeconds || 0;
      ***REMOVED***
      /**
       * @private
       * @param {Request|string***REMOVED*** request A request to run this strategy for.
       * @param {workbox-strategies.StrategyHandler***REMOVED*** handler The event that
       *     triggered the request.
       * @return {Promise<Response>***REMOVED***
       */


      async _handle(request, handler) {
        {
          assert_js.assert.isInstance(request, Request, {
            moduleName: 'workbox-strategies',
            className: this.constructor.name,
            funcName: '_handle',
            paramName: 'request'
          ***REMOVED***);
        ***REMOVED***

        let error = undefined;
        let response;

        try {
          const promises = [handler.fetch(request)];

          if (this._networkTimeoutSeconds) {
            const timeoutPromise = timeout_js.timeout(this._networkTimeoutSeconds * 1000);
            promises.push(timeoutPromise);
          ***REMOVED***

          response = await Promise.race(promises);

          if (!response) {
            throw new Error(`Timed out the network response after ` + `${this._networkTimeoutSeconds***REMOVED*** seconds.`);
          ***REMOVED***
        ***REMOVED*** catch (err) {
          if (err instanceof Error) {
            error = err;
          ***REMOVED***
        ***REMOVED***

        {
          logger_js.logger.groupCollapsed(messages.strategyStart(this.constructor.name, request));

          if (response) {
            logger_js.logger.log(`Got response from network.`);
          ***REMOVED*** else {
            logger_js.logger.log(`Unable to get a response from the network.`);
          ***REMOVED***

          messages.printFinalResponse(response);
          logger_js.logger.groupEnd();
        ***REMOVED***

        if (!response) {
          throw new WorkboxError_js.WorkboxError('no-response', {
            url: request.url,
            error
          ***REMOVED***);
        ***REMOVED***

        return response;
      ***REMOVED***

    ***REMOVED***

    /*
      Copyright 2018 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /**
     * An implementation of a
     * [stale-while-revalidate](https://developer.chrome.com/docs/workbox/caching-strategies-overview/#stale-while-revalidate)
     * request strategy.
     *
     * Resources are requested from both the cache and the network in parallel.
     * The strategy will respond with the cached version if available, otherwise
     * wait for the network response. The cache is updated with the network response
     * with each successful request.
     *
     * By default, this strategy will cache responses with a 200 status code as
     * well as [opaque responses](https://developer.chrome.com/docs/workbox/caching-resources-during-runtime/#opaque-responses).
     * Opaque responses are cross-origin requests where the response doesn't
     * support [CORS](https://enable-cors.org/).
     *
     * If the network request fails, and there is no cache match, this will throw
     * a `WorkboxError` exception.
     *
     * @extends workbox-strategies.Strategy
     * @memberof workbox-strategies
     */

    class StaleWhileRevalidate extends Strategy {
      /**
       * @param {Object***REMOVED*** [options]
       * @param {string***REMOVED*** [options.cacheName] Cache name to store and retrieve
       * requests. Defaults to cache names provided by
       * {@link workbox-core.cacheNames***REMOVED***.
       * @param {Array<Object>***REMOVED*** [options.plugins] [Plugins]{@link https://developers.google.com/web/tools/workbox/guides/using-plugins***REMOVED***
       * to use in conjunction with this caching strategy.
       * @param {Object***REMOVED*** [options.fetchOptions] Values passed along to the
       * [`init`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch#Parameters)
       * of [non-navigation](https://github.com/GoogleChrome/workbox/issues/1796)
       * `fetch()` requests made by this strategy.
       * @param {Object***REMOVED*** [options.matchOptions] [`CacheQueryOptions`](https://w3c.github.io/ServiceWorker/#dictdef-cachequeryoptions)
       */
      constructor(options = {***REMOVED***) {
        super(options); // If this instance contains no plugins with a 'cacheWillUpdate' callback,
        // prepend the `cacheOkAndOpaquePlugin` plugin to the plugins list.

        if (!this.plugins.some(p => 'cacheWillUpdate' in p)) {
          this.plugins.unshift(cacheOkAndOpaquePlugin);
        ***REMOVED***
      ***REMOVED***
      /**
       * @private
       * @param {Request|string***REMOVED*** request A request to run this strategy for.
       * @param {workbox-strategies.StrategyHandler***REMOVED*** handler The event that
       *     triggered the request.
       * @return {Promise<Response>***REMOVED***
       */


      async _handle(request, handler) {
        const logs = [];

        {
          assert_js.assert.isInstance(request, Request, {
            moduleName: 'workbox-strategies',
            className: this.constructor.name,
            funcName: 'handle',
            paramName: 'request'
          ***REMOVED***);
        ***REMOVED***

        const fetchAndCachePromise = handler.fetchAndCachePut(request).catch(() => {// Swallow this error because a 'no-response' error will be thrown in
          // main handler return flow. This will be in the `waitUntil()` flow.
        ***REMOVED***);
        void handler.waitUntil(fetchAndCachePromise);
        let response = await handler.cacheMatch(request);
        let error;

        if (response) {
          {
            logs.push(`Found a cached response in the '${this.cacheName***REMOVED***'` + ` cache. Will update with the network response in the background.`);
          ***REMOVED***
        ***REMOVED*** else {
          {
            logs.push(`No response found in the '${this.cacheName***REMOVED***' cache. ` + `Will wait for the network response.`);
          ***REMOVED***

          try {
            // NOTE(philipwalton): Really annoying that we have to type cast here.
            // https://github.com/microsoft/TypeScript/issues/20006
            response = await fetchAndCachePromise;
          ***REMOVED*** catch (err) {
            if (err instanceof Error) {
              error = err;
            ***REMOVED***
          ***REMOVED***
        ***REMOVED***

        {
          logger_js.logger.groupCollapsed(messages.strategyStart(this.constructor.name, request));

          for (const log of logs) {
            logger_js.logger.log(log);
          ***REMOVED***

          messages.printFinalResponse(response);
          logger_js.logger.groupEnd();
        ***REMOVED***

        if (!response) {
          throw new WorkboxError_js.WorkboxError('no-response', {
            url: request.url,
            error
          ***REMOVED***);
        ***REMOVED***

        return response;
      ***REMOVED***

    ***REMOVED***

    exports.CacheFirst = CacheFirst;
    exports.CacheOnly = CacheOnly;
    exports.NetworkFirst = NetworkFirst;
    exports.NetworkOnly = NetworkOnly;
    exports.StaleWhileRevalidate = StaleWhileRevalidate;
    exports.Strategy = Strategy;
    exports.StrategyHandler = StrategyHandler;

    return exports;

***REMOVED***({***REMOVED***, workbox.core._private, workbox.core._private, workbox.core._private, workbox.core._private, workbox.core._private, workbox.core._private, workbox.core._private, workbox.core._private, workbox.core._private));
