// es-abstract/2018/CompletionRecord.js
'use strict';

var $SyntaxError = require('es-errors/syntax');

var SLOT = require('internal-slot');

// https://262.ecma-international.org/7.0/#sec-completion-record-specification-type

var CompletionRecord = function CompletionRecord(type, value) {
	if (!(this instanceof CompletionRecord)) {
		return new CompletionRecord(type, value);
	***REMOVED***
	if (type !== 'normal' && type !== 'break' && type !== 'continue' && type !== 'return' && type !== 'throw') {
		throw new $SyntaxError('Assertion failed: `type` must be one of "normal", "break", "continue", "return", or "throw"');
	***REMOVED***
	SLOT.set(this, '[[Type]]', type);
	SLOT.set(this, '[[Value]]', value);
	// [[Target]] slot?
***REMOVED***;

CompletionRecord.prototype.type = function Type() {
	return SLOT.get(this, '[[Type]]');
***REMOVED***;

CompletionRecord.prototype.value = function Value() {
	return SLOT.get(this, '[[Value]]');
***REMOVED***;

CompletionRecord.prototype['?'] = function ReturnIfAbrupt() {
	var type = SLOT.get(this, '[[Type]]');
	var value = SLOT.get(this, '[[Value]]');

	if (type === 'throw') {
		throw value;
	***REMOVED***
	return value;
***REMOVED***;

CompletionRecord.prototype['!'] = function assert() {
	var type = SLOT.get(this, '[[Type]]');

	if (type !== 'normal') {
		throw new $SyntaxError('Assertion failed: Completion Record is not of type "normal"');
	***REMOVED***
	return SLOT.get(this, '[[Value]]');
***REMOVED***;

module.exports = CompletionRecord;
