// tailwindcss/lib/util/withAlphaVariable.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
***REMOVED***);
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    ***REMOVED***);
***REMOVED***
_export(exports, {
    withAlphaValue: function() {
        return withAlphaValue;
    ***REMOVED***,
    default: function() {
        return withAlphaVariable;
    ***REMOVED***
***REMOVED***);
const _color = require("./color");
function withAlphaValue(color, alphaValue, defaultValue) {
    if (typeof color === "function") {
        return color({
            opacityValue: alphaValue
        ***REMOVED***);
    ***REMOVED***
    let parsed = (0, _color.parseColor)(color, {
        loose: true
    ***REMOVED***);
    if (parsed === null) {
        return defaultValue;
    ***REMOVED***
    return (0, _color.formatColor)({
        ...parsed,
        alpha: alphaValue
    ***REMOVED***);
***REMOVED***
function withAlphaVariable({ color , property , variable  ***REMOVED***) {
    let properties = [].concat(property);
    if (typeof color === "function") {
        return {
            [variable]: "1",
            ...Object.fromEntries(properties.map((p)=>{
                return [
                    p,
                    color({
                        opacityVariable: variable,
                        opacityValue: `var(${variable***REMOVED***, 1)`
                    ***REMOVED***)
                ];
            ***REMOVED***))
        ***REMOVED***;
    ***REMOVED***
    const parsed = (0, _color.parseColor)(color);
    if (parsed === null) {
        return Object.fromEntries(properties.map((p)=>[
                p,
                color
            ]));
    ***REMOVED***
    if (parsed.alpha !== undefined) {
        // Has an alpha value, return color as-is
        return Object.fromEntries(properties.map((p)=>[
                p,
                color
            ]));
    ***REMOVED***
    return {
        [variable]: "1",
        ...Object.fromEntries(properties.map((p)=>{
            return [
                p,
                (0, _color.formatColor)({
                    ...parsed,
                    alpha: `var(${variable***REMOVED***, 1)`
                ***REMOVED***)
            ];
        ***REMOVED***))
    ***REMOVED***;
***REMOVED***
