// css-tree/lib/syntax/node/AtrulePrelude.js
var TYPE = require('../../tokenizer').TYPE;

var SEMICOLON = TYPE.Semicolon;
var LEFTCURLYBRACKET = TYPE.LeftCurlyBracket;

module.exports = {
    name: 'AtrulePrelude',
    structure: {
        children: [[]]
    ***REMOVED***,
    parse: function(name) {
        var children = null;

        if (name !== null) {
            name = name.toLowerCase();
        ***REMOVED***

        this.scanner.skipSC();

        if (this.atrule.hasOwnProperty(name) &&
            typeof this.atrule[name].prelude === 'function') {
            // custom consumer
            children = this.atrule[name].prelude.call(this);
        ***REMOVED*** else {
            // default consumer
            children = this.readSequence(this.scope.AtrulePrelude);
        ***REMOVED***

        this.scanner.skipSC();

        if (this.scanner.eof !== true &&
            this.scanner.tokenType !== LEFTCURLYBRACKET &&
            this.scanner.tokenType !== SEMICOLON) {
            this.error('Semicolon or block is expected');
        ***REMOVED***

        if (children === null) {
            children = this.createList();
        ***REMOVED***

        return {
            type: 'AtrulePrelude',
            loc: this.getLocationFromList(children),
            children: children
        ***REMOVED***;
    ***REMOVED***,
    generate: function(node) {
        this.children(node);
    ***REMOVED***,
    walkContext: 'atrulePrelude'
***REMOVED***;
