// postcss-font-variant/index.js
/**
 * font variant convertion map
 *
 * @type {Object***REMOVED***
 */
const fontVariantProperties = {
  "font-variant-ligatures": {
    "common-ligatures": "\"liga\", \"clig\"",
    "no-common-ligatures": "\"liga\", \"clig off\"",
    "discretionary-ligatures": "\"dlig\"",
    "no-discretionary-ligatures": "\"dlig\" off",
    "historical-ligatures": "\"hlig\"",
    "no-historical-ligatures": "\"hlig\" off",
    contextual: "\"calt\"",
    "no-contextual": "\"calt\" off"
  ***REMOVED***,

  "font-variant-position": {
    sub: "\"subs\"",
    "super": "\"sups\"",
    normal: "\"subs\" off, \"sups\" off"
  ***REMOVED***,

  "font-variant-caps": {
    "small-caps": "\"smcp\"",
    "all-small-caps": "\"smcp\", \"c2sc\"",
    "petite-caps": "\"pcap\"",
    "all-petite-caps": "\"pcap\", \"c2pc\"",
    unicase: "\"unic\"",
    "titling-caps": "\"titl\""
  ***REMOVED***,

  "font-variant-numeric": {
    "lining-nums": "\"lnum\"",
    "oldstyle-nums": "\"onum\"",
    "proportional-nums": "\"pnum\"",
    "tabular-nums": "\"tnum\"",
    "diagonal-fractions": "\"frac\"",
    "stacked-fractions": "\"afrc\"",
    ordinal: "\"ordn\"",
    "slashed-zero": "\"zero\""
  ***REMOVED***,

  "font-kerning": {
    normal: "\"kern\"",
    none: "\"kern\" off"
  ***REMOVED***,

  "font-variant": {
    normal: "normal",
    inherit: "inherit"
  ***REMOVED***
***REMOVED***

// The `font-variant` property is a shorthand for all the others.
for (const prop in fontVariantProperties) {
  const keys = fontVariantProperties[prop]
  for (const key in keys) {
    if (!(key in fontVariantProperties["font-variant"])) {
      fontVariantProperties["font-variant"][key] = keys[key]
    ***REMOVED***
  ***REMOVED***
***REMOVED***

// Find font-feature-settings declaration before given declaration,
// create if does not exist
function getFontFeatureSettingsPrevTo(decl) {
  let fontFeatureSettings = null;
  decl.parent.walkDecls((decl) => {
    if (decl.prop === "font-feature-settings") {
      fontFeatureSettings = decl;
    ***REMOVED***
  ***REMOVED***)

  if (fontFeatureSettings === null) {
    fontFeatureSettings = decl.clone()
    fontFeatureSettings.prop = "font-feature-settings"
    fontFeatureSettings.value = ""
    decl.parent.insertBefore(decl, fontFeatureSettings)
  ***REMOVED***
  return fontFeatureSettings
***REMOVED***

function walkRule(rule) {
  let fontFeatureSettings = null
  // read custom media queries
  rule.walkDecls((decl) => {
    if (!fontVariantProperties[decl.prop]) {
      return null
    ***REMOVED***

    let newValue = decl.value
    if (decl.prop === "font-variant") {
      newValue = decl.value.split(/\s+/g).map((val) => {
        return fontVariantProperties["font-variant"][val]
      ***REMOVED***).join(", ")
    ***REMOVED***
    else if (fontVariantProperties[decl.prop][decl.value]) {
      newValue = fontVariantProperties[decl.prop][decl.value]
    ***REMOVED***

    if (fontFeatureSettings === null) {
      fontFeatureSettings = getFontFeatureSettingsPrevTo(decl);
    ***REMOVED***
    if (fontFeatureSettings.value && fontFeatureSettings.value !== newValue) {
      fontFeatureSettings.value += ", " + newValue;
    ***REMOVED***
    else {
      fontFeatureSettings.value = newValue;
    ***REMOVED***
  ***REMOVED***)
***REMOVED***

/**
 * Expose the font-variant plugin.
 */
module.exports = () => {
  return {
    postcssPlugin: "postcss-font-variant",
    Once(root) {
      root.walkRules(walkRule)
    ***REMOVED***
  ***REMOVED***
***REMOVED***

module.exports.postcss = true
