// @babel/generator/lib/generators/base.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.BlockStatement = BlockStatement;
exports.Directive = Directive;
exports.DirectiveLiteral = DirectiveLiteral;
exports.File = File;
exports.InterpreterDirective = InterpreterDirective;
exports.Placeholder = Placeholder;
exports.Program = Program;
function File(node) {
  if (node.program) {
    this.print(node.program.interpreter);
  ***REMOVED***
  this.print(node.program);
***REMOVED***
function Program(node) {
  var _node$directives;
  this.noIndentInnerCommentsHere();
  this.printInnerComments();
  const directivesLen = (_node$directives = node.directives) == null ? void 0 : _node$directives.length;
  if (directivesLen) {
    var _node$directives$trai;
    const newline = node.body.length ? 2 : 1;
    this.printSequence(node.directives, undefined, newline);
    if (!((_node$directives$trai = node.directives[directivesLen - 1].trailingComments) != null && _node$directives$trai.length)) {
      this.newline(newline);
    ***REMOVED***
  ***REMOVED***
  this.printSequence(node.body);
***REMOVED***
function BlockStatement(node) {
  var _node$directives2;
  this.tokenChar(123);
  const exit = this.enterDelimited();
  const directivesLen = (_node$directives2 = node.directives) == null ? void 0 : _node$directives2.length;
  if (directivesLen) {
    var _node$directives$trai2;
    const newline = node.body.length ? 2 : 1;
    this.printSequence(node.directives, true, newline);
    if (!((_node$directives$trai2 = node.directives[directivesLen - 1].trailingComments) != null && _node$directives$trai2.length)) {
      this.newline(newline);
    ***REMOVED***
  ***REMOVED***
  this.printSequence(node.body, true);
  exit();
  this.rightBrace(node);
***REMOVED***
function Directive(node) {
  this.print(node.value);
  this.semicolon();
***REMOVED***
const unescapedSingleQuoteRE = /(?:^|[^\\])(?:\\\\)*'/;
const unescapedDoubleQuoteRE = /(?:^|[^\\])(?:\\\\)*"/;
function DirectiveLiteral(node) {
  const raw = this.getPossibleRaw(node);
  if (!this.format.minified && raw !== undefined) {
    this.token(raw);
    return;
  ***REMOVED***
  const {
    value
  ***REMOVED*** = node;
  if (!unescapedDoubleQuoteRE.test(value)) {
    this.token(`"${value***REMOVED***"`);
  ***REMOVED*** else if (!unescapedSingleQuoteRE.test(value)) {
    this.token(`'${value***REMOVED***'`);
  ***REMOVED*** else {
    throw new Error("Malformed AST: it is not possible to print a directive containing" + " both unescaped single and double quotes.");
  ***REMOVED***
***REMOVED***
function InterpreterDirective(node) {
  this.token(`#!${node.value***REMOVED***`);
  this.newline(1, true);
***REMOVED***
function Placeholder(node) {
  this.token("%%");
  this.print(node.name);
  this.token("%%");
  if (node.expectedNode === "Statement") {
    this.semicolon();
  ***REMOVED***
***REMOVED***
