// @babel/generator/lib/generators/expressions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.LogicalExpression = exports.BinaryExpression = exports.AssignmentExpression = AssignmentExpression;
exports.AssignmentPattern = AssignmentPattern;
exports.AwaitExpression = AwaitExpression;
exports.BindExpression = BindExpression;
exports.CallExpression = CallExpression;
exports.ConditionalExpression = ConditionalExpression;
exports.Decorator = Decorator;
exports.DoExpression = DoExpression;
exports.EmptyStatement = EmptyStatement;
exports.ExpressionStatement = ExpressionStatement;
exports.Import = Import;
exports.MemberExpression = MemberExpression;
exports.MetaProperty = MetaProperty;
exports.ModuleExpression = ModuleExpression;
exports.NewExpression = NewExpression;
exports.OptionalCallExpression = OptionalCallExpression;
exports.OptionalMemberExpression = OptionalMemberExpression;
exports.ParenthesizedExpression = ParenthesizedExpression;
exports.PrivateName = PrivateName;
exports.SequenceExpression = SequenceExpression;
exports.Super = Super;
exports.ThisExpression = ThisExpression;
exports.UnaryExpression = UnaryExpression;
exports.UpdateExpression = UpdateExpression;
exports.V8IntrinsicIdentifier = V8IntrinsicIdentifier;
exports.YieldExpression = YieldExpression;
exports._shouldPrintDecoratorsBeforeExport = _shouldPrintDecoratorsBeforeExport;
var _t = require("@babel/types");
var _index = require("../node/index.js");
const {
  isCallExpression,
  isLiteral,
  isMemberExpression,
  isNewExpression,
  isPattern
***REMOVED*** = _t;
function UnaryExpression(node) {
  const {
    operator
  ***REMOVED*** = node;
  if (operator === "void" || operator === "delete" || operator === "typeof" || operator === "throw") {
    this.word(operator);
    this.space();
  ***REMOVED*** else {
    this.token(operator);
  ***REMOVED***
  this.print(node.argument);
***REMOVED***
function DoExpression(node) {
  if (node.async) {
    this.word("async", true);
    this.space();
  ***REMOVED***
  this.word("do");
  this.space();
  this.print(node.body);
***REMOVED***
function ParenthesizedExpression(node) {
  this.tokenChar(40);
  const exit = this.enterDelimited();
  this.print(node.expression);
  exit();
  this.rightParens(node);
***REMOVED***
function UpdateExpression(node) {
  if (node.prefix) {
    this.token(node.operator);
    this.print(node.argument);
  ***REMOVED*** else {
    this.print(node.argument, true);
    this.token(node.operator);
  ***REMOVED***
***REMOVED***
function ConditionalExpression(node) {
  this.print(node.test);
  this.space();
  this.tokenChar(63);
  this.space();
  this.print(node.consequent);
  this.space();
  this.tokenChar(58);
  this.space();
  this.print(node.alternate);
***REMOVED***
function NewExpression(node, parent) {
  this.word("new");
  this.space();
  this.print(node.callee);
  if (this.format.minified && node.arguments.length === 0 && !node.optional && !isCallExpression(parent, {
    callee: node
  ***REMOVED***) && !isMemberExpression(parent) && !isNewExpression(parent)) {
    return;
  ***REMOVED***
  this.print(node.typeArguments);
  {
    this.print(node.typeParameters);
  ***REMOVED***
  if (node.optional) {
    this.token("?.");
  ***REMOVED***
  if (node.arguments.length === 0 && this.tokenMap && !this.tokenMap.endMatches(node, ")")) {
    return;
  ***REMOVED***
  this.tokenChar(40);
  const exit = this.enterDelimited();
  this.printList(node.arguments, this.shouldPrintTrailingComma(")"));
  exit();
  this.rightParens(node);
***REMOVED***
function SequenceExpression(node) {
  this.printList(node.expressions);
***REMOVED***
function ThisExpression() {
  this.word("this");
***REMOVED***
function Super() {
  this.word("super");
***REMOVED***
function _shouldPrintDecoratorsBeforeExport(node) {
  if (typeof this.format.decoratorsBeforeExport === "boolean") {
    return this.format.decoratorsBeforeExport;
  ***REMOVED***
  return typeof node.start === "number" && node.start === node.declaration.start;
***REMOVED***
function Decorator(node) {
  this.tokenChar(64);
  this.print(node.expression);
  this.newline();
***REMOVED***
function OptionalMemberExpression(node) {
  let {
    computed
  ***REMOVED*** = node;
  const {
    optional,
    property
  ***REMOVED*** = node;
  this.print(node.object);
  if (!computed && isMemberExpression(property)) {
    throw new TypeError("Got a MemberExpression for MemberExpression property");
  ***REMOVED***
  if (isLiteral(property) && typeof property.value === "number") {
    computed = true;
  ***REMOVED***
  if (optional) {
    this.token("?.");
  ***REMOVED***
  if (computed) {
    this.tokenChar(91);
    this.print(property);
    this.tokenChar(93);
  ***REMOVED*** else {
    if (!optional) {
      this.tokenChar(46);
    ***REMOVED***
    this.print(property);
  ***REMOVED***
***REMOVED***
function OptionalCallExpression(node) {
  this.print(node.callee);
  {
    this.print(node.typeParameters);
  ***REMOVED***
  if (node.optional) {
    this.token("?.");
  ***REMOVED***
  this.print(node.typeArguments);
  this.tokenChar(40);
  const exit = this.enterDelimited();
  this.printList(node.arguments);
  exit();
  this.rightParens(node);
***REMOVED***
function CallExpression(node) {
  this.print(node.callee);
  this.print(node.typeArguments);
  {
    this.print(node.typeParameters);
  ***REMOVED***
  this.tokenChar(40);
  const exit = this.enterDelimited();
  this.printList(node.arguments, this.shouldPrintTrailingComma(")"));
  exit();
  this.rightParens(node);
***REMOVED***
function Import() {
  this.word("import");
***REMOVED***
function AwaitExpression(node) {
  this.word("await");
  if (node.argument) {
    this.space();
    this.printTerminatorless(node.argument);
  ***REMOVED***
***REMOVED***
function YieldExpression(node) {
  this.word("yield", true);
  if (node.delegate) {
    this.tokenChar(42);
    if (node.argument) {
      this.space();
      this.print(node.argument);
    ***REMOVED***
  ***REMOVED*** else {
    if (node.argument) {
      this.space();
      this.printTerminatorless(node.argument);
    ***REMOVED***
  ***REMOVED***
***REMOVED***
function EmptyStatement() {
  this.semicolon(true);
***REMOVED***
function ExpressionStatement(node) {
  this.tokenContext |= _index.TokenContext.expressionStatement;
  this.print(node.expression);
  this.semicolon();
***REMOVED***
function AssignmentPattern(node) {
  this.print(node.left);
  if (node.left.type === "Identifier" || isPattern(node.left)) {
    if (node.left.optional) this.tokenChar(63);
    this.print(node.left.typeAnnotation);
  ***REMOVED***
  this.space();
  this.tokenChar(61);
  this.space();
  this.print(node.right);
***REMOVED***
function AssignmentExpression(node) {
  this.print(node.left);
  this.space();
  if (node.operator === "in" || node.operator === "instanceof") {
    this.word(node.operator);
  ***REMOVED*** else {
    this.token(node.operator);
    this._endsWithDiv = node.operator === "/";
  ***REMOVED***
  this.space();
  this.print(node.right);
***REMOVED***
function BindExpression(node) {
  this.print(node.object);
  this.token("::");
  this.print(node.callee);
***REMOVED***
function MemberExpression(node) {
  this.print(node.object);
  if (!node.computed && isMemberExpression(node.property)) {
    throw new TypeError("Got a MemberExpression for MemberExpression property");
  ***REMOVED***
  let computed = node.computed;
  if (isLiteral(node.property) && typeof node.property.value === "number") {
    computed = true;
  ***REMOVED***
  if (computed) {
    const exit = this.enterDelimited();
    this.tokenChar(91);
    this.print(node.property);
    this.tokenChar(93);
    exit();
  ***REMOVED*** else {
    this.tokenChar(46);
    this.print(node.property);
  ***REMOVED***
***REMOVED***
function MetaProperty(node) {
  this.print(node.meta);
  this.tokenChar(46);
  this.print(node.property);
***REMOVED***
function PrivateName(node) {
  this.tokenChar(35);
  this.print(node.id);
***REMOVED***
function V8IntrinsicIdentifier(node) {
  this.tokenChar(37);
  this.word(node.name);
***REMOVED***
function ModuleExpression(node) {
  this.word("module", true);
  this.space();
  this.tokenChar(123);
  this.indent();
  const {
    body
  ***REMOVED*** = node;
  if (body.body.length || body.directives.length) {
    this.newline();
  ***REMOVED***
  this.print(body);
  this.dedent();
  this.rightBrace(node);
***REMOVED***
