// @babel/generator/lib/generators/statements.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.BreakStatement = BreakStatement;
exports.CatchClause = CatchClause;
exports.ContinueStatement = ContinueStatement;
exports.DebuggerStatement = DebuggerStatement;
exports.DoWhileStatement = DoWhileStatement;
exports.ForOfStatement = exports.ForInStatement = void 0;
exports.ForStatement = ForStatement;
exports.IfStatement = IfStatement;
exports.LabeledStatement = LabeledStatement;
exports.ReturnStatement = ReturnStatement;
exports.SwitchCase = SwitchCase;
exports.SwitchStatement = SwitchStatement;
exports.ThrowStatement = ThrowStatement;
exports.TryStatement = TryStatement;
exports.VariableDeclaration = VariableDeclaration;
exports.VariableDeclarator = VariableDeclarator;
exports.WhileStatement = WhileStatement;
exports.WithStatement = WithStatement;
var _t = require("@babel/types");
var _index = require("../node/index.js");
const {
  isFor,
  isForStatement,
  isIfStatement,
  isStatement
***REMOVED*** = _t;
function WithStatement(node) {
  this.word("with");
  this.space();
  this.tokenChar(40);
  this.print(node.object);
  this.tokenChar(41);
  this.printBlock(node);
***REMOVED***
function IfStatement(node) {
  this.word("if");
  this.space();
  this.tokenChar(40);
  this.print(node.test);
  this.tokenChar(41);
  this.space();
  const needsBlock = node.alternate && isIfStatement(getLastStatement(node.consequent));
  if (needsBlock) {
    this.tokenChar(123);
    this.newline();
    this.indent();
  ***REMOVED***
  this.printAndIndentOnComments(node.consequent);
  if (needsBlock) {
    this.dedent();
    this.newline();
    this.tokenChar(125);
  ***REMOVED***
  if (node.alternate) {
    if (this.endsWith(125)) this.space();
    this.word("else");
    this.space();
    this.printAndIndentOnComments(node.alternate);
  ***REMOVED***
***REMOVED***
function getLastStatement(statement) {
  const {
    body
  ***REMOVED*** = statement;
  if (isStatement(body) === false) {
    return statement;
  ***REMOVED***
  return getLastStatement(body);
***REMOVED***
function ForStatement(node) {
  this.word("for");
  this.space();
  this.tokenChar(40);
  {
    const exit = this.enterForStatementInit();
    this.tokenContext |= _index.TokenContext.forHead;
    this.print(node.init);
    exit();
  ***REMOVED***
  this.tokenChar(59);
  if (node.test) {
    this.space();
    this.print(node.test);
  ***REMOVED***
  this.token(";", false, 1);
  if (node.update) {
    this.space();
    this.print(node.update);
  ***REMOVED***
  this.tokenChar(41);
  this.printBlock(node);
***REMOVED***
function WhileStatement(node) {
  this.word("while");
  this.space();
  this.tokenChar(40);
  this.print(node.test);
  this.tokenChar(41);
  this.printBlock(node);
***REMOVED***
function ForXStatement(node) {
  this.word("for");
  this.space();
  const isForOf = node.type === "ForOfStatement";
  if (isForOf && node.await) {
    this.word("await");
    this.space();
  ***REMOVED***
  this.noIndentInnerCommentsHere();
  this.tokenChar(40);
  {
    const exit = isForOf ? null : this.enterForStatementInit();
    this.tokenContext |= isForOf ? _index.TokenContext.forOfHead : _index.TokenContext.forInHead;
    this.print(node.left);
    exit == null || exit();
  ***REMOVED***
  this.space();
  this.word(isForOf ? "of" : "in");
  this.space();
  this.print(node.right);
  this.tokenChar(41);
  this.printBlock(node);
***REMOVED***
const ForInStatement = exports.ForInStatement = ForXStatement;
const ForOfStatement = exports.ForOfStatement = ForXStatement;
function DoWhileStatement(node) {
  this.word("do");
  this.space();
  this.print(node.body);
  this.space();
  this.word("while");
  this.space();
  this.tokenChar(40);
  this.print(node.test);
  this.tokenChar(41);
  this.semicolon();
***REMOVED***
function printStatementAfterKeyword(printer, node) {
  if (node) {
    printer.space();
    printer.printTerminatorless(node);
  ***REMOVED***
  printer.semicolon();
***REMOVED***
function BreakStatement(node) {
  this.word("break");
  printStatementAfterKeyword(this, node.label);
***REMOVED***
function ContinueStatement(node) {
  this.word("continue");
  printStatementAfterKeyword(this, node.label);
***REMOVED***
function ReturnStatement(node) {
  this.word("return");
  printStatementAfterKeyword(this, node.argument);
***REMOVED***
function ThrowStatement(node) {
  this.word("throw");
  printStatementAfterKeyword(this, node.argument);
***REMOVED***
function LabeledStatement(node) {
  this.print(node.label);
  this.tokenChar(58);
  this.space();
  this.print(node.body);
***REMOVED***
function TryStatement(node) {
  this.word("try");
  this.space();
  this.print(node.block);
  this.space();
  if (node.handlers) {
    this.print(node.handlers[0]);
  ***REMOVED*** else {
    this.print(node.handler);
  ***REMOVED***
  if (node.finalizer) {
    this.space();
    this.word("finally");
    this.space();
    this.print(node.finalizer);
  ***REMOVED***
***REMOVED***
function CatchClause(node) {
  this.word("catch");
  this.space();
  if (node.param) {
    this.tokenChar(40);
    this.print(node.param);
    this.print(node.param.typeAnnotation);
    this.tokenChar(41);
    this.space();
  ***REMOVED***
  this.print(node.body);
***REMOVED***
function SwitchStatement(node) {
  this.word("switch");
  this.space();
  this.tokenChar(40);
  this.print(node.discriminant);
  this.tokenChar(41);
  this.space();
  this.tokenChar(123);
  this.printSequence(node.cases, true, undefined, function addNewlines(leading, cas) {
    if (!leading && node.cases[node.cases.length - 1] === cas) return -1;
  ***REMOVED***);
  this.rightBrace(node);
***REMOVED***
function SwitchCase(node) {
  if (node.test) {
    this.word("case");
    this.space();
    this.print(node.test);
    this.tokenChar(58);
  ***REMOVED*** else {
    this.word("default");
    this.tokenChar(58);
  ***REMOVED***
  if (node.consequent.length) {
    this.newline();
    this.printSequence(node.consequent, true);
  ***REMOVED***
***REMOVED***
function DebuggerStatement() {
  this.word("debugger");
  this.semicolon();
***REMOVED***
function VariableDeclaration(node, parent) {
  if (node.declare) {
    this.word("declare");
    this.space();
  ***REMOVED***
  const {
    kind
  ***REMOVED*** = node;
  if (kind === "await using") {
    this.word("await");
    this.space();
    this.word("using", true);
  ***REMOVED*** else {
    this.word(kind, kind === "using");
  ***REMOVED***
  this.space();
  let hasInits = false;
  if (!isFor(parent)) {
    for (const declar of node.declarations) {
      if (declar.init) {
        hasInits = true;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  this.printList(node.declarations, undefined, undefined, node.declarations.length > 1, hasInits ? function (occurrenceCount) {
    this.token(",", false, occurrenceCount);
    this.newline();
  ***REMOVED*** : undefined);
  if (isFor(parent)) {
    if (isForStatement(parent)) {
      if (parent.init === node) return;
    ***REMOVED*** else {
      if (parent.left === node) return;
    ***REMOVED***
  ***REMOVED***
  this.semicolon();
***REMOVED***
function VariableDeclarator(node) {
  this.print(node.id);
  if (node.definite) this.tokenChar(33);
  this.print(node.id.typeAnnotation);
  if (node.init) {
    this.space();
    this.tokenChar(61);
    this.space();
    this.print(node.init);
  ***REMOVED***
***REMOVED***
