// @babel/generator/lib/generators/typescript.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.TSAnyKeyword = TSAnyKeyword;
exports.TSArrayType = TSArrayType;
exports.TSSatisfiesExpression = exports.TSAsExpression = TSTypeExpression;
exports.TSBigIntKeyword = TSBigIntKeyword;
exports.TSBooleanKeyword = TSBooleanKeyword;
exports.TSCallSignatureDeclaration = TSCallSignatureDeclaration;
exports.TSInterfaceHeritage = exports.TSClassImplements = TSClassImplements;
exports.TSConditionalType = TSConditionalType;
exports.TSConstructSignatureDeclaration = TSConstructSignatureDeclaration;
exports.TSConstructorType = TSConstructorType;
exports.TSDeclareFunction = TSDeclareFunction;
exports.TSDeclareMethod = TSDeclareMethod;
exports.TSEnumBody = TSEnumBody;
exports.TSEnumDeclaration = TSEnumDeclaration;
exports.TSEnumMember = TSEnumMember;
exports.TSExportAssignment = TSExportAssignment;
exports.TSExternalModuleReference = TSExternalModuleReference;
exports.TSFunctionType = TSFunctionType;
exports.TSImportEqualsDeclaration = TSImportEqualsDeclaration;
exports.TSImportType = TSImportType;
exports.TSIndexSignature = TSIndexSignature;
exports.TSIndexedAccessType = TSIndexedAccessType;
exports.TSInferType = TSInferType;
exports.TSInstantiationExpression = TSInstantiationExpression;
exports.TSInterfaceBody = TSInterfaceBody;
exports.TSInterfaceDeclaration = TSInterfaceDeclaration;
exports.TSIntersectionType = TSIntersectionType;
exports.TSIntrinsicKeyword = TSIntrinsicKeyword;
exports.TSLiteralType = TSLiteralType;
exports.TSMappedType = TSMappedType;
exports.TSMethodSignature = TSMethodSignature;
exports.TSModuleBlock = TSModuleBlock;
exports.TSModuleDeclaration = TSModuleDeclaration;
exports.TSNamedTupleMember = TSNamedTupleMember;
exports.TSNamespaceExportDeclaration = TSNamespaceExportDeclaration;
exports.TSNeverKeyword = TSNeverKeyword;
exports.TSNonNullExpression = TSNonNullExpression;
exports.TSNullKeyword = TSNullKeyword;
exports.TSNumberKeyword = TSNumberKeyword;
exports.TSObjectKeyword = TSObjectKeyword;
exports.TSOptionalType = TSOptionalType;
exports.TSParameterProperty = TSParameterProperty;
exports.TSParenthesizedType = TSParenthesizedType;
exports.TSPropertySignature = TSPropertySignature;
exports.TSQualifiedName = TSQualifiedName;
exports.TSRestType = TSRestType;
exports.TSStringKeyword = TSStringKeyword;
exports.TSSymbolKeyword = TSSymbolKeyword;
exports.TSTemplateLiteralType = TSTemplateLiteralType;
exports.TSThisType = TSThisType;
exports.TSTupleType = TSTupleType;
exports.TSTypeAliasDeclaration = TSTypeAliasDeclaration;
exports.TSTypeAnnotation = TSTypeAnnotation;
exports.TSTypeAssertion = TSTypeAssertion;
exports.TSTypeLiteral = TSTypeLiteral;
exports.TSTypeOperator = TSTypeOperator;
exports.TSTypeParameter = TSTypeParameter;
exports.TSTypeParameterDeclaration = exports.TSTypeParameterInstantiation = TSTypeParameterInstantiation;
exports.TSTypePredicate = TSTypePredicate;
exports.TSTypeQuery = TSTypeQuery;
exports.TSTypeReference = TSTypeReference;
exports.TSUndefinedKeyword = TSUndefinedKeyword;
exports.TSUnionType = TSUnionType;
exports.TSUnknownKeyword = TSUnknownKeyword;
exports.TSVoidKeyword = TSVoidKeyword;
exports.tsPrintClassMemberModifiers = tsPrintClassMemberModifiers;
exports.tsPrintFunctionOrConstructorType = tsPrintFunctionOrConstructorType;
exports.tsPrintPropertyOrMethodName = tsPrintPropertyOrMethodName;
exports.tsPrintSignatureDeclarationBase = tsPrintSignatureDeclarationBase;
function TSTypeAnnotation(node, parent) {
  this.token((parent.type === "TSFunctionType" || parent.type === "TSConstructorType") && parent.typeAnnotation === node ? "=>" : ":");
  this.space();
  if (node.optional) this.tokenChar(63);
  this.print(node.typeAnnotation);
***REMOVED***
function TSTypeParameterInstantiation(node, parent) {
  this.tokenChar(60);
  let printTrailingSeparator = parent.type === "ArrowFunctionExpression" && node.params.length === 1;
  if (this.tokenMap && node.start != null && node.end != null) {
    printTrailingSeparator && (printTrailingSeparator = !!this.tokenMap.find(node, t => this.tokenMap.matchesOriginal(t, ",")));
    printTrailingSeparator || (printTrailingSeparator = this.shouldPrintTrailingComma(">"));
  ***REMOVED***
  this.printList(node.params, printTrailingSeparator);
  this.tokenChar(62);
***REMOVED***
function TSTypeParameter(node) {
  if (node.const) {
    this.word("const");
    this.space();
  ***REMOVED***
  if (node.in) {
    this.word("in");
    this.space();
  ***REMOVED***
  if (node.out) {
    this.word("out");
    this.space();
  ***REMOVED***
  this.word(node.name);
  if (node.constraint) {
    this.space();
    this.word("extends");
    this.space();
    this.print(node.constraint);
  ***REMOVED***
  if (node.default) {
    this.space();
    this.tokenChar(61);
    this.space();
    this.print(node.default);
  ***REMOVED***
***REMOVED***
function TSParameterProperty(node) {
  if (node.accessibility) {
    this.word(node.accessibility);
    this.space();
  ***REMOVED***
  if (node.readonly) {
    this.word("readonly");
    this.space();
  ***REMOVED***
  this._param(node.parameter);
***REMOVED***
function TSDeclareFunction(node, parent) {
  if (node.declare) {
    this.word("declare");
    this.space();
  ***REMOVED***
  this._functionHead(node, parent);
  this.semicolon();
***REMOVED***
function TSDeclareMethod(node) {
  this._classMethodHead(node);
  this.semicolon();
***REMOVED***
function TSQualifiedName(node) {
  this.print(node.left);
  this.tokenChar(46);
  this.print(node.right);
***REMOVED***
function TSCallSignatureDeclaration(node) {
  this.tsPrintSignatureDeclarationBase(node);
  maybePrintTrailingCommaOrSemicolon(this, node);
***REMOVED***
function maybePrintTrailingCommaOrSemicolon(printer, node) {
  if (!printer.tokenMap || !node.start || !node.end) {
    printer.semicolon();
    return;
  ***REMOVED***
  if (printer.tokenMap.endMatches(node, ",")) {
    printer.token(",");
  ***REMOVED*** else if (printer.tokenMap.endMatches(node, ";")) {
    printer.semicolon();
  ***REMOVED***
***REMOVED***
function TSConstructSignatureDeclaration(node) {
  this.word("new");
  this.space();
  this.tsPrintSignatureDeclarationBase(node);
  maybePrintTrailingCommaOrSemicolon(this, node);
***REMOVED***
function TSPropertySignature(node) {
  const {
    readonly
  ***REMOVED*** = node;
  if (readonly) {
    this.word("readonly");
    this.space();
  ***REMOVED***
  this.tsPrintPropertyOrMethodName(node);
  this.print(node.typeAnnotation);
  maybePrintTrailingCommaOrSemicolon(this, node);
***REMOVED***
function tsPrintPropertyOrMethodName(node) {
  if (node.computed) {
    this.tokenChar(91);
  ***REMOVED***
  this.print(node.key);
  if (node.computed) {
    this.tokenChar(93);
  ***REMOVED***
  if (node.optional) {
    this.tokenChar(63);
  ***REMOVED***
***REMOVED***
function TSMethodSignature(node) {
  const {
    kind
  ***REMOVED*** = node;
  if (kind === "set" || kind === "get") {
    this.word(kind);
    this.space();
  ***REMOVED***
  this.tsPrintPropertyOrMethodName(node);
  this.tsPrintSignatureDeclarationBase(node);
  maybePrintTrailingCommaOrSemicolon(this, node);
***REMOVED***
function TSIndexSignature(node) {
  const {
    readonly,
    static: isStatic
  ***REMOVED*** = node;
  if (isStatic) {
    this.word("static");
    this.space();
  ***REMOVED***
  if (readonly) {
    this.word("readonly");
    this.space();
  ***REMOVED***
  this.tokenChar(91);
  this._parameters(node.parameters, "]");
  this.print(node.typeAnnotation);
  maybePrintTrailingCommaOrSemicolon(this, node);
***REMOVED***
function TSAnyKeyword() {
  this.word("any");
***REMOVED***
function TSBigIntKeyword() {
  this.word("bigint");
***REMOVED***
function TSUnknownKeyword() {
  this.word("unknown");
***REMOVED***
function TSNumberKeyword() {
  this.word("number");
***REMOVED***
function TSObjectKeyword() {
  this.word("object");
***REMOVED***
function TSBooleanKeyword() {
  this.word("boolean");
***REMOVED***
function TSStringKeyword() {
  this.word("string");
***REMOVED***
function TSSymbolKeyword() {
  this.word("symbol");
***REMOVED***
function TSVoidKeyword() {
  this.word("void");
***REMOVED***
function TSUndefinedKeyword() {
  this.word("undefined");
***REMOVED***
function TSNullKeyword() {
  this.word("null");
***REMOVED***
function TSNeverKeyword() {
  this.word("never");
***REMOVED***
function TSIntrinsicKeyword() {
  this.word("intrinsic");
***REMOVED***
function TSThisType() {
  this.word("this");
***REMOVED***
function TSFunctionType(node) {
  this.tsPrintFunctionOrConstructorType(node);
***REMOVED***
function TSConstructorType(node) {
  if (node.abstract) {
    this.word("abstract");
    this.space();
  ***REMOVED***
  this.word("new");
  this.space();
  this.tsPrintFunctionOrConstructorType(node);
***REMOVED***
function tsPrintFunctionOrConstructorType(node) {
  const {
    typeParameters
  ***REMOVED*** = node;
  const parameters = node.parameters;
  this.print(typeParameters);
  this.tokenChar(40);
  this._parameters(parameters, ")");
  this.space();
  const returnType = node.typeAnnotation;
  this.print(returnType);
***REMOVED***
function TSTypeReference(node) {
  const typeArguments = node.typeParameters;
  this.print(node.typeName, !!typeArguments);
  this.print(typeArguments);
***REMOVED***
function TSTypePredicate(node) {
  if (node.asserts) {
    this.word("asserts");
    this.space();
  ***REMOVED***
  this.print(node.parameterName);
  if (node.typeAnnotation) {
    this.space();
    this.word("is");
    this.space();
    this.print(node.typeAnnotation.typeAnnotation);
  ***REMOVED***
***REMOVED***
function TSTypeQuery(node) {
  this.word("typeof");
  this.space();
  this.print(node.exprName);
  const typeArguments = node.typeParameters;
  if (typeArguments) {
    this.print(typeArguments);
  ***REMOVED***
***REMOVED***
function TSTypeLiteral(node) {
  printBraced(this, node, () => this.printJoin(node.members, true, true));
***REMOVED***
function TSArrayType(node) {
  this.print(node.elementType, true);
  this.tokenChar(91);
  this.tokenChar(93);
***REMOVED***
function TSTupleType(node) {
  this.tokenChar(91);
  this.printList(node.elementTypes, this.shouldPrintTrailingComma("]"));
  this.tokenChar(93);
***REMOVED***
function TSOptionalType(node) {
  this.print(node.typeAnnotation);
  this.tokenChar(63);
***REMOVED***
function TSRestType(node) {
  this.token("...");
  this.print(node.typeAnnotation);
***REMOVED***
function TSNamedTupleMember(node) {
  this.print(node.label);
  if (node.optional) this.tokenChar(63);
  this.tokenChar(58);
  this.space();
  this.print(node.elementType);
***REMOVED***
function TSUnionType(node) {
  tsPrintUnionOrIntersectionType(this, node, "|");
***REMOVED***
function TSIntersectionType(node) {
  tsPrintUnionOrIntersectionType(this, node, "&");
***REMOVED***
function tsPrintUnionOrIntersectionType(printer, node, sep) {
  var _printer$tokenMap;
  let hasLeadingToken = 0;
  if ((_printer$tokenMap = printer.tokenMap) != null && _printer$tokenMap.startMatches(node, sep)) {
    hasLeadingToken = 1;
    printer.token(sep);
  ***REMOVED***
  printer.printJoin(node.types, undefined, undefined, function (i) {
    this.space();
    this.token(sep, null, i + hasLeadingToken);
    this.space();
  ***REMOVED***);
***REMOVED***
function TSConditionalType(node) {
  this.print(node.checkType);
  this.space();
  this.word("extends");
  this.space();
  this.print(node.extendsType);
  this.space();
  this.tokenChar(63);
  this.space();
  this.print(node.trueType);
  this.space();
  this.tokenChar(58);
  this.space();
  this.print(node.falseType);
***REMOVED***
function TSInferType(node) {
  this.word("infer");
  this.print(node.typeParameter);
***REMOVED***
function TSParenthesizedType(node) {
  this.tokenChar(40);
  this.print(node.typeAnnotation);
  this.tokenChar(41);
***REMOVED***
function TSTypeOperator(node) {
  this.word(node.operator);
  this.space();
  this.print(node.typeAnnotation);
***REMOVED***
function TSIndexedAccessType(node) {
  this.print(node.objectType, true);
  this.tokenChar(91);
  this.print(node.indexType);
  this.tokenChar(93);
***REMOVED***
function TSMappedType(node) {
  const {
    nameType,
    optional,
    readonly,
    typeAnnotation
  ***REMOVED*** = node;
  this.tokenChar(123);
  const exit = this.enterDelimited();
  this.space();
  if (readonly) {
    tokenIfPlusMinus(this, readonly);
    this.word("readonly");
    this.space();
  ***REMOVED***
  this.tokenChar(91);
  {
    this.word(node.typeParameter.name);
  ***REMOVED***
  this.space();
  this.word("in");
  this.space();
  {
    this.print(node.typeParameter.constraint);
  ***REMOVED***
  if (nameType) {
    this.space();
    this.word("as");
    this.space();
    this.print(nameType);
  ***REMOVED***
  this.tokenChar(93);
  if (optional) {
    tokenIfPlusMinus(this, optional);
    this.tokenChar(63);
  ***REMOVED***
  if (typeAnnotation) {
    this.tokenChar(58);
    this.space();
    this.print(typeAnnotation);
  ***REMOVED***
  this.space();
  exit();
  this.tokenChar(125);
***REMOVED***
function tokenIfPlusMinus(self, tok) {
  if (tok !== true) {
    self.token(tok);
  ***REMOVED***
***REMOVED***
function TSTemplateLiteralType(node) {
  this._printTemplate(node, node.types);
***REMOVED***
function TSLiteralType(node) {
  this.print(node.literal);
***REMOVED***
function TSClassImplements(node) {
  this.print(node.expression);
  this.print(node.typeArguments);
***REMOVED***
function TSInterfaceDeclaration(node) {
  const {
    declare,
    id,
    typeParameters,
    extends: extendz,
    body
  ***REMOVED*** = node;
  if (declare) {
    this.word("declare");
    this.space();
  ***REMOVED***
  this.word("interface");
  this.space();
  this.print(id);
  this.print(typeParameters);
  if (extendz != null && extendz.length) {
    this.space();
    this.word("extends");
    this.space();
    this.printList(extendz);
  ***REMOVED***
  this.space();
  this.print(body);
***REMOVED***
function TSInterfaceBody(node) {
  printBraced(this, node, () => this.printJoin(node.body, true, true));
***REMOVED***
function TSTypeAliasDeclaration(node) {
  const {
    declare,
    id,
    typeParameters,
    typeAnnotation
  ***REMOVED*** = node;
  if (declare) {
    this.word("declare");
    this.space();
  ***REMOVED***
  this.word("type");
  this.space();
  this.print(id);
  this.print(typeParameters);
  this.space();
  this.tokenChar(61);
  this.space();
  this.print(typeAnnotation);
  this.semicolon();
***REMOVED***
function TSTypeExpression(node) {
  const {
    type,
    expression,
    typeAnnotation
  ***REMOVED*** = node;
  this.print(expression, true);
  this.space();
  this.word(type === "TSAsExpression" ? "as" : "satisfies");
  this.space();
  this.print(typeAnnotation);
***REMOVED***
function TSTypeAssertion(node) {
  const {
    typeAnnotation,
    expression
  ***REMOVED*** = node;
  this.tokenChar(60);
  this.print(typeAnnotation);
  this.tokenChar(62);
  this.space();
  this.print(expression);
***REMOVED***
function TSInstantiationExpression(node) {
  this.print(node.expression);
  {
    this.print(node.typeParameters);
  ***REMOVED***
***REMOVED***
function TSEnumDeclaration(node) {
  const {
    declare,
    const: isConst,
    id
  ***REMOVED*** = node;
  if (declare) {
    this.word("declare");
    this.space();
  ***REMOVED***
  if (isConst) {
    this.word("const");
    this.space();
  ***REMOVED***
  this.word("enum");
  this.space();
  this.print(id);
  this.space();
  {
    TSEnumBody.call(this, node);
  ***REMOVED***
***REMOVED***
function TSEnumBody(node) {
  printBraced(this, node, () => {
    var _this$shouldPrintTrai;
    return this.printList(node.members, (_this$shouldPrintTrai = this.shouldPrintTrailingComma("***REMOVED***")) != null ? _this$shouldPrintTrai : true, true, true);
  ***REMOVED***);
***REMOVED***
function TSEnumMember(node) {
  const {
    id,
    initializer
  ***REMOVED*** = node;
  this.print(id);
  if (initializer) {
    this.space();
    this.tokenChar(61);
    this.space();
    this.print(initializer);
  ***REMOVED***
***REMOVED***
function TSModuleDeclaration(node) {
  const {
    declare,
    id,
    kind
  ***REMOVED*** = node;
  if (declare) {
    this.word("declare");
    this.space();
  ***REMOVED***
  {
    if (!node.global) {
      this.word(kind != null ? kind : id.type === "Identifier" ? "namespace" : "module");
      this.space();
    ***REMOVED***
    this.print(id);
    if (!node.body) {
      this.semicolon();
      return;
    ***REMOVED***
    let body = node.body;
    while (body.type === "TSModuleDeclaration") {
      this.tokenChar(46);
      this.print(body.id);
      body = body.body;
    ***REMOVED***
    this.space();
    this.print(body);
  ***REMOVED***
***REMOVED***
function TSModuleBlock(node) {
  printBraced(this, node, () => this.printSequence(node.body, true));
***REMOVED***
function TSImportType(node) {
  const {
    argument,
    qualifier,
    options
  ***REMOVED*** = node;
  this.word("import");
  this.tokenChar(40);
  this.print(argument);
  if (options) {
    this.tokenChar(44);
    this.print(options);
  ***REMOVED***
  this.tokenChar(41);
  if (qualifier) {
    this.tokenChar(46);
    this.print(qualifier);
  ***REMOVED***
  const typeArguments = node.typeParameters;
  if (typeArguments) {
    this.print(typeArguments);
  ***REMOVED***
***REMOVED***
function TSImportEqualsDeclaration(node) {
  const {
    id,
    moduleReference
  ***REMOVED*** = node;
  if (node.isExport) {
    this.word("export");
    this.space();
  ***REMOVED***
  this.word("import");
  this.space();
  this.print(id);
  this.space();
  this.tokenChar(61);
  this.space();
  this.print(moduleReference);
  this.semicolon();
***REMOVED***
function TSExternalModuleReference(node) {
  this.token("require(");
  this.print(node.expression);
  this.tokenChar(41);
***REMOVED***
function TSNonNullExpression(node) {
  this.print(node.expression);
  this.tokenChar(33);
***REMOVED***
function TSExportAssignment(node) {
  this.word("export");
  this.space();
  this.tokenChar(61);
  this.space();
  this.print(node.expression);
  this.semicolon();
***REMOVED***
function TSNamespaceExportDeclaration(node) {
  this.word("export");
  this.space();
  this.word("as");
  this.space();
  this.word("namespace");
  this.space();
  this.print(node.id);
  this.semicolon();
***REMOVED***
function tsPrintSignatureDeclarationBase(node) {
  const {
    typeParameters
  ***REMOVED*** = node;
  const parameters = node.parameters;
  this.print(typeParameters);
  this.tokenChar(40);
  this._parameters(parameters, ")");
  const returnType = node.typeAnnotation;
  this.print(returnType);
***REMOVED***
function tsPrintClassMemberModifiers(node) {
  const isPrivateField = node.type === "ClassPrivateProperty";
  const isPublicField = node.type === "ClassAccessorProperty" || node.type === "ClassProperty";
  printModifiersList(this, node, [isPublicField && node.declare && "declare", !isPrivateField && node.accessibility]);
  if (node.static) {
    this.word("static");
    this.space();
  ***REMOVED***
  printModifiersList(this, node, [!isPrivateField && node.abstract && "abstract", !isPrivateField && node.override && "override", (isPublicField || isPrivateField) && node.readonly && "readonly"]);
***REMOVED***
function printBraced(printer, node, cb) {
  printer.token("{");
  const exit = printer.enterDelimited();
  cb();
  exit();
  printer.rightBrace(node);
***REMOVED***
function printModifiersList(printer, node, modifiers) {
  var _printer$tokenMap2;
  const modifiersSet = new Set();
  for (const modifier of modifiers) {
    if (modifier) modifiersSet.add(modifier);
  ***REMOVED***
  (_printer$tokenMap2 = printer.tokenMap) == null || _printer$tokenMap2.find(node, tok => {
    if (modifiersSet.has(tok.value)) {
      printer.token(tok.value);
      printer.space();
      modifiersSet.delete(tok.value);
      return modifiersSet.size === 0;
    ***REMOVED***
  ***REMOVED***);
  for (const modifier of modifiersSet) {
    printer.word(modifier);
    printer.space();
  ***REMOVED***
***REMOVED***
