// css-tree/lib/generator/create.js
var sourceMap = require('./sourceMap');
var hasOwnProperty = Object.prototype.hasOwnProperty;

function processChildren(node, delimeter) {
    var list = node.children;
    var prev = null;

    if (typeof delimeter !== 'function') {
        list.forEach(this.node, this);
    ***REMOVED*** else {
        list.forEach(function(node) {
            if (prev !== null) {
                delimeter.call(this, prev);
            ***REMOVED***

            this.node(node);
            prev = node;
        ***REMOVED***, this);
    ***REMOVED***
***REMOVED***

module.exports = function createGenerator(config) {
    function processNode(node) {
        if (hasOwnProperty.call(types, node.type)) {
            types[node.type].call(this, node);
        ***REMOVED*** else {
            throw new Error('Unknown node type: ' + node.type);
        ***REMOVED***
    ***REMOVED***

    var types = {***REMOVED***;

    if (config.node) {
        for (var name in config.node) {
            types[name] = config.node[name].generate;
        ***REMOVED***
    ***REMOVED***

    return function(node, options) {
        var buffer = '';
        var handlers = {
            children: processChildren,
            node: processNode,
            chunk: function(chunk) {
                buffer += chunk;
            ***REMOVED***,
            result: function() {
                return buffer;
            ***REMOVED***
        ***REMOVED***;

        if (options) {
            if (typeof options.decorator === 'function') {
                handlers = options.decorator(handlers);
            ***REMOVED***

            if (options.sourceMap) {
                handlers = sourceMap(handlers);
            ***REMOVED***
        ***REMOVED***

        handlers.node(node);

        return handlers.result();
    ***REMOVED***;
***REMOVED***;
