// co/index.js

/**
 * slice() reference.
 */

var slice = Array.prototype.slice;

/**
 * Expose `co`.
 */

module.exports = co['default'] = co.co = co;

/**
 * Wrap the given generator `fn` into a
 * function that returns a promise.
 * This is a separate function so that
 * every `co()` call doesn't create a new,
 * unnecessary closure.
 *
 * @param {GeneratorFunction***REMOVED*** fn
 * @return {Function***REMOVED***
 * @api public
 */

co.wrap = function (fn) {
  createPromise.__generatorFunction__ = fn;
  return createPromise;
  function createPromise() {
    return co.call(this, fn.apply(this, arguments));
  ***REMOVED***
***REMOVED***;

/**
 * Execute the generator function or a generator
 * and return a promise.
 *
 * @param {Function***REMOVED*** fn
 * @return {Promise***REMOVED***
 * @api public
 */

function co(gen) {
  var ctx = this;
  var args = slice.call(arguments, 1)

  // we wrap everything in a promise to avoid promise chaining,
  // which leads to memory leak errors.
  // see https://github.com/tj/co/issues/180
  return new Promise(function(resolve, reject) {
    if (typeof gen === 'function') gen = gen.apply(ctx, args);
    if (!gen || typeof gen.next !== 'function') return resolve(gen);

    onFulfilled();

    /**
     * @param {Mixed***REMOVED*** res
     * @return {Promise***REMOVED***
     * @api private
     */

    function onFulfilled(res) {
      var ret;
      try {
        ret = gen.next(res);
      ***REMOVED*** catch (e) {
        return reject(e);
      ***REMOVED***
      next(ret);
    ***REMOVED***

    /**
     * @param {Error***REMOVED*** err
     * @return {Promise***REMOVED***
     * @api private
     */

    function onRejected(err) {
      var ret;
      try {
        ret = gen.throw(err);
      ***REMOVED*** catch (e) {
        return reject(e);
      ***REMOVED***
      next(ret);
    ***REMOVED***

    /**
     * Get the next value in the generator,
     * return a promise.
     *
     * @param {Object***REMOVED*** ret
     * @return {Promise***REMOVED***
     * @api private
     */

    function next(ret) {
      if (ret.done) return resolve(ret.value);
      var value = toPromise.call(ctx, ret.value);
      if (value && isPromise(value)) return value.then(onFulfilled, onRejected);
      return onRejected(new TypeError('You may only yield a function, promise, generator, array, or object, '
        + 'but the following object was passed: "' + String(ret.value) + '"'));
    ***REMOVED***
  ***REMOVED***);
***REMOVED***

/**
 * Convert a `yield`ed value into a promise.
 *
 * @param {Mixed***REMOVED*** obj
 * @return {Promise***REMOVED***
 * @api private
 */

function toPromise(obj) {
  if (!obj) return obj;
  if (isPromise(obj)) return obj;
  if (isGeneratorFunction(obj) || isGenerator(obj)) return co.call(this, obj);
  if ('function' == typeof obj) return thunkToPromise.call(this, obj);
  if (Array.isArray(obj)) return arrayToPromise.call(this, obj);
  if (isObject(obj)) return objectToPromise.call(this, obj);
  return obj;
***REMOVED***

/**
 * Convert a thunk to a promise.
 *
 * @param {Function***REMOVED***
 * @return {Promise***REMOVED***
 * @api private
 */

function thunkToPromise(fn) {
  var ctx = this;
  return new Promise(function (resolve, reject) {
    fn.call(ctx, function (err, res) {
      if (err) return reject(err);
      if (arguments.length > 2) res = slice.call(arguments, 1);
      resolve(res);
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***

/**
 * Convert an array of "yieldables" to a promise.
 * Uses `Promise.all()` internally.
 *
 * @param {Array***REMOVED*** obj
 * @return {Promise***REMOVED***
 * @api private
 */

function arrayToPromise(obj) {
  return Promise.all(obj.map(toPromise, this));
***REMOVED***

/**
 * Convert an object of "yieldables" to a promise.
 * Uses `Promise.all()` internally.
 *
 * @param {Object***REMOVED*** obj
 * @return {Promise***REMOVED***
 * @api private
 */

function objectToPromise(obj){
  var results = new obj.constructor();
  var keys = Object.keys(obj);
  var promises = [];
  for (var i = 0; i < keys.length; i++) {
    var key = keys[i];
    var promise = toPromise.call(this, obj[key]);
    if (promise && isPromise(promise)) defer(promise, key);
    else results[key] = obj[key];
  ***REMOVED***
  return Promise.all(promises).then(function () {
    return results;
  ***REMOVED***);

  function defer(promise, key) {
    // predefine the key in the result
    results[key] = undefined;
    promises.push(promise.then(function (res) {
      results[key] = res;
    ***REMOVED***));
  ***REMOVED***
***REMOVED***

/**
 * Check if `obj` is a promise.
 *
 * @param {Object***REMOVED*** obj
 * @return {Boolean***REMOVED***
 * @api private
 */

function isPromise(obj) {
  return 'function' == typeof obj.then;
***REMOVED***

/**
 * Check if `obj` is a generator.
 *
 * @param {Mixed***REMOVED*** obj
 * @return {Boolean***REMOVED***
 * @api private
 */

function isGenerator(obj) {
  return 'function' == typeof obj.next && 'function' == typeof obj.throw;
***REMOVED***

/**
 * Check if `obj` is a generator function.
 *
 * @param {Mixed***REMOVED*** obj
 * @return {Boolean***REMOVED***
 * @api private
 */
function isGeneratorFunction(obj) {
  var constructor = obj.constructor;
  if (!constructor) return false;
  if ('GeneratorFunction' === constructor.name || 'GeneratorFunction' === constructor.displayName) return true;
  return isGenerator(constructor.prototype);
***REMOVED***

/**
 * Check for plain object.
 *
 * @param {Mixed***REMOVED*** val
 * @return {Boolean***REMOVED***
 * @api private
 */

function isObject(val) {
  return Object == val.constructor;
***REMOVED***
