// @pmmmwh/react-refresh-webpack-plugin/overlay/theme.js
/**
 * @typedef {Object***REMOVED*** Theme
 * @property {string[]***REMOVED*** reset
 * @property {string***REMOVED*** black
 * @property {string***REMOVED*** red
 * @property {string***REMOVED*** green
 * @property {string***REMOVED*** yellow
 * @property {string***REMOVED*** blue
 * @property {string***REMOVED*** magenta
 * @property {string***REMOVED*** cyan
 * @property {string***REMOVED*** white
 * @property {string***REMOVED*** lightgrey
 * @property {string***REMOVED*** darkgrey
 * @property {string***REMOVED*** grey
 * @property {string***REMOVED*** dimgrey
 */

/**
 * @type {Theme***REMOVED*** theme
 * A collection of colors to be used by the overlay.
 * Partially adopted from Tomorrow Night Bright.
 */
const theme = {
  reset: ['transparent', 'transparent'],
  black: '000000',
  red: 'D34F56',
  green: 'B9C954',
  yellow: 'E6C452',
  blue: '7CA7D8',
  magenta: 'C299D6',
  cyan: '73BFB1',
  white: 'FFFFFF',
  lightgrey: 'C7C7C7',
  darkgrey: 'A9A9A9',
  grey: '474747',
  dimgrey: '343434',
***REMOVED***;

module.exports = theme;
