// @eslint-community/eslint-utils/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true ***REMOVED***);

var eslintVisitorKeys = require('eslint-visitor-keys');

/** @typedef {import("eslint").Scope.Scope***REMOVED*** Scope */
/** @typedef {import("estree").Node***REMOVED*** Node */

/**
 * Get the innermost scope which contains a given location.
 * @param {Scope***REMOVED*** initialScope The initial scope to search.
 * @param {Node***REMOVED*** node The location to search.
 * @returns {Scope***REMOVED*** The innermost scope.
 */
function getInnermostScope(initialScope, node) {
    const location = /** @type {[number, number]***REMOVED*** */ (node.range)[0];

    let scope = initialScope;
    let found = false;
    do {
        found = false;
        for (const childScope of scope.childScopes) {
            const range = /** @type {[number, number]***REMOVED*** */ (
                childScope.block.range
            );

            if (range[0] <= location && location < range[1]) {
                scope = childScope;
                found = true;
                break
            ***REMOVED***
        ***REMOVED***
    ***REMOVED*** while (found)

    return scope
***REMOVED***

/** @typedef {import("eslint").Scope.Scope***REMOVED*** Scope */
/** @typedef {import("eslint").Scope.Variable***REMOVED*** Variable */
/** @typedef {import("estree").Identifier***REMOVED*** Identifier */

/**
 * Find the variable of a given name.
 * @param {Scope***REMOVED*** initialScope The scope to start finding.
 * @param {string|Identifier***REMOVED*** nameOrNode The variable name to find. If this is a Node object then it should be an Identifier node.
 * @returns {Variable|null***REMOVED*** The found variable or null.
 */
function findVariable(initialScope, nameOrNode) {
    let name = "";
    /** @type {Scope|null***REMOVED*** */
    let scope = initialScope;

    if (typeof nameOrNode === "string") {
        name = nameOrNode;
    ***REMOVED*** else {
        name = nameOrNode.name;
        scope = getInnermostScope(scope, nameOrNode);
    ***REMOVED***

    while (scope != null) {
        const variable = scope.set.get(name);
        if (variable != null) {
            return variable
        ***REMOVED***
        scope = scope.upper;
    ***REMOVED***

    return null
***REMOVED***

/** @typedef {import("eslint").AST.Token***REMOVED*** Token */
/** @typedef {import("estree").Comment***REMOVED*** Comment */
/** @typedef {import("./types.mjs").ArrowToken***REMOVED*** ArrowToken */
/** @typedef {import("./types.mjs").CommaToken***REMOVED*** CommaToken */
/** @typedef {import("./types.mjs").SemicolonToken***REMOVED*** SemicolonToken */
/** @typedef {import("./types.mjs").ColonToken***REMOVED*** ColonToken */
/** @typedef {import("./types.mjs").OpeningParenToken***REMOVED*** OpeningParenToken */
/** @typedef {import("./types.mjs").ClosingParenToken***REMOVED*** ClosingParenToken */
/** @typedef {import("./types.mjs").OpeningBracketToken***REMOVED*** OpeningBracketToken */
/** @typedef {import("./types.mjs").ClosingBracketToken***REMOVED*** ClosingBracketToken */
/** @typedef {import("./types.mjs").OpeningBraceToken***REMOVED*** OpeningBraceToken */
/** @typedef {import("./types.mjs").ClosingBraceToken***REMOVED*** ClosingBraceToken */
/**
 * @template {string***REMOVED*** Value
 * @typedef {import("./types.mjs").PunctuatorToken<Value>***REMOVED*** PunctuatorToken
 */

/** @typedef {Comment | Token***REMOVED*** CommentOrToken */

/**
 * Creates the negate function of the given function.
 * @param {function(CommentOrToken):boolean***REMOVED*** f - The function to negate.
 * @returns {function(CommentOrToken):boolean***REMOVED*** Negated function.
 */
function negate(f) {
    return (token) => !f(token)
***REMOVED***

/**
 * Checks if the given token is a PunctuatorToken with the given value
 * @template {string***REMOVED*** Value
 * @param {CommentOrToken***REMOVED*** token - The token to check.
 * @param {Value***REMOVED*** value - The value to check.
 * @returns {token is PunctuatorToken<Value>***REMOVED*** `true` if the token is a PunctuatorToken with the given value.
 */
function isPunctuatorTokenWithValue(token, value) {
    return token.type === "Punctuator" && token.value === value
***REMOVED***

/**
 * Checks if the given token is an arrow token or not.
 * @param {CommentOrToken***REMOVED*** token - The token to check.
 * @returns {token is ArrowToken***REMOVED*** `true` if the token is an arrow token.
 */
function isArrowToken(token) {
    return isPunctuatorTokenWithValue(token, "=>")
***REMOVED***

/**
 * Checks if the given token is a comma token or not.
 * @param {CommentOrToken***REMOVED*** token - The token to check.
 * @returns {token is CommaToken***REMOVED*** `true` if the token is a comma token.
 */
function isCommaToken(token) {
    return isPunctuatorTokenWithValue(token, ",")
***REMOVED***

/**
 * Checks if the given token is a semicolon token or not.
 * @param {CommentOrToken***REMOVED*** token - The token to check.
 * @returns {token is SemicolonToken***REMOVED*** `true` if the token is a semicolon token.
 */
function isSemicolonToken(token) {
    return isPunctuatorTokenWithValue(token, ";")
***REMOVED***

/**
 * Checks if the given token is a colon token or not.
 * @param {CommentOrToken***REMOVED*** token - The token to check.
 * @returns {token is ColonToken***REMOVED*** `true` if the token is a colon token.
 */
function isColonToken(token) {
    return isPunctuatorTokenWithValue(token, ":")
***REMOVED***

/**
 * Checks if the given token is an opening parenthesis token or not.
 * @param {CommentOrToken***REMOVED*** token - The token to check.
 * @returns {token is OpeningParenToken***REMOVED*** `true` if the token is an opening parenthesis token.
 */
function isOpeningParenToken(token) {
    return isPunctuatorTokenWithValue(token, "(")
***REMOVED***

/**
 * Checks if the given token is a closing parenthesis token or not.
 * @param {CommentOrToken***REMOVED*** token - The token to check.
 * @returns {token is ClosingParenToken***REMOVED*** `true` if the token is a closing parenthesis token.
 */
function isClosingParenToken(token) {
    return isPunctuatorTokenWithValue(token, ")")
***REMOVED***

/**
 * Checks if the given token is an opening square bracket token or not.
 * @param {CommentOrToken***REMOVED*** token - The token to check.
 * @returns {token is OpeningBracketToken***REMOVED*** `true` if the token is an opening square bracket token.
 */
function isOpeningBracketToken(token) {
    return isPunctuatorTokenWithValue(token, "[")
***REMOVED***

/**
 * Checks if the given token is a closing square bracket token or not.
 * @param {CommentOrToken***REMOVED*** token - The token to check.
 * @returns {token is ClosingBracketToken***REMOVED*** `true` if the token is a closing square bracket token.
 */
function isClosingBracketToken(token) {
    return isPunctuatorTokenWithValue(token, "]")
***REMOVED***

/**
 * Checks if the given token is an opening brace token or not.
 * @param {CommentOrToken***REMOVED*** token - The token to check.
 * @returns {token is OpeningBraceToken***REMOVED*** `true` if the token is an opening brace token.
 */
function isOpeningBraceToken(token) {
    return isPunctuatorTokenWithValue(token, "{")
***REMOVED***

/**
 * Checks if the given token is a closing brace token or not.
 * @param {CommentOrToken***REMOVED*** token - The token to check.
 * @returns {token is ClosingBraceToken***REMOVED*** `true` if the token is a closing brace token.
 */
function isClosingBraceToken(token) {
    return isPunctuatorTokenWithValue(token, "***REMOVED***")
***REMOVED***

/**
 * Checks if the given token is a comment token or not.
 * @param {CommentOrToken***REMOVED*** token - The token to check.
 * @returns {token is Comment***REMOVED*** `true` if the token is a comment token.
 */
function isCommentToken(token) {
    return ["Block", "Line", "Shebang"].includes(token.type)
***REMOVED***

const isNotArrowToken = negate(isArrowToken);
const isNotCommaToken = negate(isCommaToken);
const isNotSemicolonToken = negate(isSemicolonToken);
const isNotColonToken = negate(isColonToken);
const isNotOpeningParenToken = negate(isOpeningParenToken);
const isNotClosingParenToken = negate(isClosingParenToken);
const isNotOpeningBracketToken = negate(isOpeningBracketToken);
const isNotClosingBracketToken = negate(isClosingBracketToken);
const isNotOpeningBraceToken = negate(isOpeningBraceToken);
const isNotClosingBraceToken = negate(isClosingBraceToken);
const isNotCommentToken = negate(isCommentToken);

/** @typedef {import("eslint").Rule.Node***REMOVED*** RuleNode */
/** @typedef {import("eslint").SourceCode***REMOVED*** SourceCode */
/** @typedef {import("eslint").AST.Token***REMOVED*** Token */
/** @typedef {import("estree").Function***REMOVED*** FunctionNode */
/** @typedef {import("estree").FunctionDeclaration***REMOVED*** FunctionDeclaration */
/** @typedef {import("estree").FunctionExpression***REMOVED*** FunctionExpression */
/** @typedef {import("estree").SourceLocation***REMOVED*** SourceLocation */
/** @typedef {import("estree").Position***REMOVED*** Position */

/**
 * Get the `(` token of the given function node.
 * @param {FunctionExpression | FunctionDeclaration***REMOVED*** node - The function node to get.
 * @param {SourceCode***REMOVED*** sourceCode - The source code object to get tokens.
 * @returns {Token***REMOVED*** `(` token.
 */
function getOpeningParenOfParams(node, sourceCode) {
    return node.id
        ? /** @type {Token***REMOVED*** */ (
              sourceCode.getTokenAfter(node.id, isOpeningParenToken)
          )
        : /** @type {Token***REMOVED*** */ (
              sourceCode.getFirstToken(node, isOpeningParenToken)
          )
***REMOVED***

/**
 * Get the location of the given function node for reporting.
 * @param {FunctionNode***REMOVED*** node - The function node to get.
 * @param {SourceCode***REMOVED*** sourceCode - The source code object to get tokens.
 * @returns {SourceLocation|null***REMOVED*** The location of the function node for reporting.
 */
function getFunctionHeadLocation(node, sourceCode) {
    const parent = /** @type {RuleNode***REMOVED*** */ (node).parent;

    /** @type {Position|null***REMOVED*** */
    let start = null;
    /** @type {Position|null***REMOVED*** */
    let end = null;

    if (node.type === "ArrowFunctionExpression") {
        const arrowToken = /** @type {Token***REMOVED*** */ (
            sourceCode.getTokenBefore(node.body, isArrowToken)
        );

        start = arrowToken.loc.start;
        end = arrowToken.loc.end;
    ***REMOVED*** else if (
        parent.type === "Property" ||
        parent.type === "MethodDefinition" ||
        parent.type === "PropertyDefinition"
    ) {
        start = /** @type {SourceLocation***REMOVED*** */ (parent.loc).start;
        end = getOpeningParenOfParams(node, sourceCode).loc.start;
    ***REMOVED*** else {
        start = /** @type {SourceLocation***REMOVED*** */ (node.loc).start;
        end = getOpeningParenOfParams(node, sourceCode).loc.start;
    ***REMOVED***

    return {
        start: { ...start ***REMOVED***,
        end: { ...end ***REMOVED***,
    ***REMOVED***
***REMOVED***

/* globals globalThis, global, self, window */
/** @typedef {import("./types.mjs").StaticValue***REMOVED*** StaticValue */
/** @typedef {import("eslint").Scope.Scope***REMOVED*** Scope */
/** @typedef {import("estree").Node***REMOVED*** Node */
/** @typedef {import("@typescript-eslint/types").TSESTree.Node***REMOVED*** TSESTreeNode */
/** @typedef {import("@typescript-eslint/types").TSESTree.AST_NODE_TYPES***REMOVED*** TSESTreeNodeTypes */
/** @typedef {import("@typescript-eslint/types").TSESTree.MemberExpression***REMOVED*** MemberExpression */
/** @typedef {import("@typescript-eslint/types").TSESTree.Property***REMOVED*** Property */
/** @typedef {import("@typescript-eslint/types").TSESTree.RegExpLiteral***REMOVED*** RegExpLiteral */
/** @typedef {import("@typescript-eslint/types").TSESTree.BigIntLiteral***REMOVED*** BigIntLiteral */
/** @typedef {import("@typescript-eslint/types").TSESTree.Literal***REMOVED*** Literal */

const globalObject =
    typeof globalThis !== "undefined"
        ? globalThis
        : // @ts-ignore
        typeof self !== "undefined"
        ? // @ts-ignore
          self
        : // @ts-ignore
        typeof window !== "undefined"
        ? // @ts-ignore
          window
        : typeof global !== "undefined"
        ? global
        : {***REMOVED***;

const builtinNames = Object.freeze(
    new Set([
        "Array",
        "ArrayBuffer",
        "BigInt",
        "BigInt64Array",
        "BigUint64Array",
        "Boolean",
        "DataView",
        "Date",
        "decodeURI",
        "decodeURIComponent",
        "encodeURI",
        "encodeURIComponent",
        "escape",
        "Float32Array",
        "Float64Array",
        "Function",
        "Infinity",
        "Int16Array",
        "Int32Array",
        "Int8Array",
        "isFinite",
        "isNaN",
        "isPrototypeOf",
        "JSON",
        "Map",
        "Math",
        "NaN",
        "Number",
        "Object",
        "parseFloat",
        "parseInt",
        "Promise",
        "Proxy",
        "Reflect",
        "RegExp",
        "Set",
        "String",
        "Symbol",
        "Uint16Array",
        "Uint32Array",
        "Uint8Array",
        "Uint8ClampedArray",
        "undefined",
        "unescape",
        "WeakMap",
        "WeakSet",
    ]),
);
const callAllowed = new Set(
    [
        Array.isArray,
        Array.of,
        Array.prototype.at,
        Array.prototype.concat,
        Array.prototype.entries,
        Array.prototype.every,
        Array.prototype.filter,
        Array.prototype.find,
        Array.prototype.findIndex,
        Array.prototype.flat,
        Array.prototype.includes,
        Array.prototype.indexOf,
        Array.prototype.join,
        Array.prototype.keys,
        Array.prototype.lastIndexOf,
        Array.prototype.slice,
        Array.prototype.some,
        Array.prototype.toString,
        Array.prototype.values,
        typeof BigInt === "function" ? BigInt : undefined,
        Boolean,
        Date,
        Date.parse,
        decodeURI,
        decodeURIComponent,
        encodeURI,
        encodeURIComponent,
        escape,
        isFinite,
        isNaN,
        // @ts-ignore
        isPrototypeOf,
        Map,
        Map.prototype.entries,
        Map.prototype.get,
        Map.prototype.has,
        Map.prototype.keys,
        Map.prototype.values,
        .../** @type {(keyof typeof Math)[]***REMOVED*** */ (
            Object.getOwnPropertyNames(Math)
        )
            .filter((k) => k !== "random")
            .map((k) => Math[k])
            .filter((f) => typeof f === "function"),
        Number,
        Number.isFinite,
        Number.isNaN,
        Number.parseFloat,
        Number.parseInt,
        Number.prototype.toExponential,
        Number.prototype.toFixed,
        Number.prototype.toPrecision,
        Number.prototype.toString,
        Object,
        Object.entries,
        Object.is,
        Object.isExtensible,
        Object.isFrozen,
        Object.isSealed,
        Object.keys,
        Object.values,
        parseFloat,
        parseInt,
        RegExp,
        Set,
        Set.prototype.entries,
        Set.prototype.has,
        Set.prototype.keys,
        Set.prototype.values,
        String,
        String.fromCharCode,
        String.fromCodePoint,
        String.raw,
        String.prototype.at,
        String.prototype.charAt,
        String.prototype.charCodeAt,
        String.prototype.codePointAt,
        String.prototype.concat,
        String.prototype.endsWith,
        String.prototype.includes,
        String.prototype.indexOf,
        String.prototype.lastIndexOf,
        String.prototype.normalize,
        String.prototype.padEnd,
        String.prototype.padStart,
        String.prototype.slice,
        String.prototype.startsWith,
        String.prototype.substr,
        String.prototype.substring,
        String.prototype.toLowerCase,
        String.prototype.toString,
        String.prototype.toUpperCase,
        String.prototype.trim,
        String.prototype.trimEnd,
        String.prototype.trimLeft,
        String.prototype.trimRight,
        String.prototype.trimStart,
        Symbol.for,
        Symbol.keyFor,
        unescape,
    ].filter((f) => typeof f === "function"),
);
const callPassThrough = new Set([
    Object.freeze,
    Object.preventExtensions,
    Object.seal,
]);

/** @type {ReadonlyArray<readonly [Function, ReadonlySet<string>]>***REMOVED*** */
const getterAllowed = [
    [Map, new Set(["size"])],
    [
        RegExp,
        new Set([
            "dotAll",
            "flags",
            "global",
            "hasIndices",
            "ignoreCase",
            "multiline",
            "source",
            "sticky",
            "unicode",
        ]),
    ],
    [Set, new Set(["size"])],
];

/**
 * Get the property descriptor.
 * @param {object***REMOVED*** object The object to get.
 * @param {string|number|symbol***REMOVED*** name The property name to get.
 */
function getPropertyDescriptor(object, name) {
    let x = object;
    while ((typeof x === "object" || typeof x === "function") && x !== null) {
        const d = Object.getOwnPropertyDescriptor(x, name);
        if (d) {
            return d
        ***REMOVED***
        x = Object.getPrototypeOf(x);
    ***REMOVED***
    return null
***REMOVED***

/**
 * Check if a property is getter or not.
 * @param {object***REMOVED*** object The object to check.
 * @param {string|number|symbol***REMOVED*** name The property name to check.
 */
function isGetter(object, name) {
    const d = getPropertyDescriptor(object, name);
    return d != null && d.get != null
***REMOVED***

/**
 * Get the element values of a given node list.
 * @param {(Node|TSESTreeNode|null)[]***REMOVED*** nodeList The node list to get values.
 * @param {Scope|undefined|null***REMOVED*** initialScope The initial scope to find variables.
 * @returns {any[]|null***REMOVED*** The value list if all nodes are constant. Otherwise, null.
 */
function getElementValues(nodeList, initialScope) {
    const valueList = [];

    for (let i = 0; i < nodeList.length; ++i) {
        const elementNode = nodeList[i];

        if (elementNode == null) {
            valueList.length = i + 1;
        ***REMOVED*** else if (elementNode.type === "SpreadElement") {
            const argument = getStaticValueR(elementNode.argument, initialScope);
            if (argument == null) {
                return null
            ***REMOVED***
            valueList.push(.../** @type {Iterable<any>***REMOVED*** */ (argument.value));
        ***REMOVED*** else {
            const element = getStaticValueR(elementNode, initialScope);
            if (element == null) {
                return null
            ***REMOVED***
            valueList.push(element.value);
        ***REMOVED***
    ***REMOVED***

    return valueList
***REMOVED***

/**
 * Returns whether the given variable is never written to after initialization.
 * @param {import("eslint").Scope.Variable***REMOVED*** variable
 * @returns {boolean***REMOVED***
 */
function isEffectivelyConst(variable) {
    const refs = variable.references;

    const inits = refs.filter((r) => r.init).length;
    const reads = refs.filter((r) => r.isReadOnly()).length;
    if (inits === 1 && reads + inits === refs.length) {
        // there is only one init and all other references only read
        return true
    ***REMOVED***
    return false
***REMOVED***

/**
 * @template {TSESTreeNodeTypes***REMOVED*** T
 * @callback VisitorCallback
 * @param {TSESTreeNode & { type: T ***REMOVED******REMOVED*** node
 * @param {Scope|undefined|null***REMOVED*** initialScope
 * @returns {StaticValue | null***REMOVED***
 */
/**
 * @typedef { { [K in TSESTreeNodeTypes]?: VisitorCallback<K> ***REMOVED*** ***REMOVED*** Operations
 */
/**
 * @type {Operations***REMOVED***
 */
const operations = Object.freeze({
    ArrayExpression(node, initialScope) {
        const elements = getElementValues(node.elements, initialScope);
        return elements != null ? { value: elements ***REMOVED*** : null
    ***REMOVED***,

    AssignmentExpression(node, initialScope) {
        if (node.operator === "=") {
            return getStaticValueR(node.right, initialScope)
        ***REMOVED***
        return null
    ***REMOVED***,

    //eslint-disable-next-line complexity
    BinaryExpression(node, initialScope) {
        if (node.operator === "in" || node.operator === "instanceof") {
            // Not supported.
            return null
        ***REMOVED***

        const left = getStaticValueR(node.left, initialScope);
        const right = getStaticValueR(node.right, initialScope);
        if (left != null && right != null) {
            switch (node.operator) {
                case "==":
                    return { value: left.value == right.value ***REMOVED*** //eslint-disable-line eqeqeq
                case "!=":
                    return { value: left.value != right.value ***REMOVED*** //eslint-disable-line eqeqeq
                case "===":
                    return { value: left.value === right.value ***REMOVED***
                case "!==":
                    return { value: left.value !== right.value ***REMOVED***
                case "<":
                    return {
                        value:
                            /** @type {any***REMOVED*** */ (left.value) <
                            /** @type {any***REMOVED*** */ (right.value),
                    ***REMOVED***
                case "<=":
                    return {
                        value:
                            /** @type {any***REMOVED*** */ (left.value) <=
                            /** @type {any***REMOVED*** */ (right.value),
                    ***REMOVED***
                case ">":
                    return {
                        value:
                            /** @type {any***REMOVED*** */ (left.value) >
                            /** @type {any***REMOVED*** */ (right.value),
                    ***REMOVED***
                case ">=":
                    return {
                        value:
                            /** @type {any***REMOVED*** */ (left.value) >=
                            /** @type {any***REMOVED*** */ (right.value),
                    ***REMOVED***
                case "<<":
                    return {
                        value:
                            /** @type {any***REMOVED*** */ (left.value) <<
                            /** @type {any***REMOVED*** */ (right.value),
                    ***REMOVED***
                case ">>":
                    return {
                        value:
                            /** @type {any***REMOVED*** */ (left.value) >>
                            /** @type {any***REMOVED*** */ (right.value),
                    ***REMOVED***
                case ">>>":
                    return {
                        value:
                            /** @type {any***REMOVED*** */ (left.value) >>>
                            /** @type {any***REMOVED*** */ (right.value),
                    ***REMOVED***
                case "+":
                    return {
                        value:
                            /** @type {any***REMOVED*** */ (left.value) +
                            /** @type {any***REMOVED*** */ (right.value),
                    ***REMOVED***
                case "-":
                    return {
                        value:
                            /** @type {any***REMOVED*** */ (left.value) -
                            /** @type {any***REMOVED*** */ (right.value),
                    ***REMOVED***
                case "*":
                    return {
                        value:
                            /** @type {any***REMOVED*** */ (left.value) *
                            /** @type {any***REMOVED*** */ (right.value),
                    ***REMOVED***
                case "/":
                    return {
                        value:
                            /** @type {any***REMOVED*** */ (left.value) /
                            /** @type {any***REMOVED*** */ (right.value),
                    ***REMOVED***
                case "%":
                    return {
                        value:
                            /** @type {any***REMOVED*** */ (left.value) %
                            /** @type {any***REMOVED*** */ (right.value),
                    ***REMOVED***
                case "**":
                    return {
                        value:
                            /** @type {any***REMOVED*** */ (left.value) **
                            /** @type {any***REMOVED*** */ (right.value),
                    ***REMOVED***
                case "|":
                    return {
                        value:
                            /** @type {any***REMOVED*** */ (left.value) |
                            /** @type {any***REMOVED*** */ (right.value),
                    ***REMOVED***
                case "^":
                    return {
                        value:
                            /** @type {any***REMOVED*** */ (left.value) ^
                            /** @type {any***REMOVED*** */ (right.value),
                    ***REMOVED***
                case "&":
                    return {
                        value:
                            /** @type {any***REMOVED*** */ (left.value) &
                            /** @type {any***REMOVED*** */ (right.value),
                    ***REMOVED***

                // no default
            ***REMOVED***
        ***REMOVED***

        return null
    ***REMOVED***,

    CallExpression(node, initialScope) {
        const calleeNode = node.callee;
        const args = getElementValues(node.arguments, initialScope);

        if (args != null) {
            if (calleeNode.type === "MemberExpression") {
                if (calleeNode.property.type === "PrivateIdentifier") {
                    return null
                ***REMOVED***
                const object = getStaticValueR(calleeNode.object, initialScope);
                if (object != null) {
                    if (
                        object.value == null &&
                        (object.optional || node.optional)
                    ) {
                        return { value: undefined, optional: true ***REMOVED***
                    ***REMOVED***
                    const property = getStaticPropertyNameValue(
                        calleeNode,
                        initialScope,
                    );

                    if (property != null) {
                        const receiver =
                            /** @type {Record<PropertyKey, (...args: any[]) => any>***REMOVED*** */ (
                                object.value
                            );
                        const methodName = /** @type {PropertyKey***REMOVED*** */ (
                            property.value
                        );
                        if (callAllowed.has(receiver[methodName])) {
                            return {
                                value: receiver[methodName](...args),
                            ***REMOVED***
                        ***REMOVED***
                        if (callPassThrough.has(receiver[methodName])) {
                            return { value: args[0] ***REMOVED***
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED*** else {
                const callee = getStaticValueR(calleeNode, initialScope);
                if (callee != null) {
                    if (callee.value == null && node.optional) {
                        return { value: undefined, optional: true ***REMOVED***
                    ***REMOVED***
                    const func = /** @type {(...args: any[]) => any***REMOVED*** */ (
                        callee.value
                    );
                    if (callAllowed.has(func)) {
                        return { value: func(...args) ***REMOVED***
                    ***REMOVED***
                    if (callPassThrough.has(func)) {
                        return { value: args[0] ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        return null
    ***REMOVED***,

    ConditionalExpression(node, initialScope) {
        const test = getStaticValueR(node.test, initialScope);
        if (test != null) {
            return test.value
                ? getStaticValueR(node.consequent, initialScope)
                : getStaticValueR(node.alternate, initialScope)
        ***REMOVED***
        return null
    ***REMOVED***,

    ExpressionStatement(node, initialScope) {
        return getStaticValueR(node.expression, initialScope)
    ***REMOVED***,

    Identifier(node, initialScope) {
        if (initialScope != null) {
            const variable = findVariable(initialScope, node);

            // Built-in globals.
            if (
                variable != null &&
                variable.defs.length === 0 &&
                builtinNames.has(variable.name) &&
                variable.name in globalObject
            ) {
                return { value: globalObject[variable.name] ***REMOVED***
            ***REMOVED***

            // Constants.
            if (variable != null && variable.defs.length === 1) {
                const def = variable.defs[0];
                if (
                    def.parent &&
                    def.type === "Variable" &&
                    (def.parent.kind === "const" ||
                        isEffectivelyConst(variable)) &&
                    // TODO(mysticatea): don't support destructuring here.
                    def.node.id.type === "Identifier"
                ) {
                    return getStaticValueR(def.node.init, initialScope)
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        return null
    ***REMOVED***,

    Literal(node) {
        const literal =
            /** @type {Partial<Literal> & Partial<RegExpLiteral> & Partial<BigIntLiteral>***REMOVED*** */ (
                node
            );
        //istanbul ignore if : this is implementation-specific behavior.
        if (
            (literal.regex != null || literal.bigint != null) &&
            literal.value == null
        ) {
            // It was a RegExp/BigInt literal, but Node.js didn't support it.
            return null
        ***REMOVED***
        return { value: literal.value ***REMOVED***
    ***REMOVED***,

    LogicalExpression(node, initialScope) {
        const left = getStaticValueR(node.left, initialScope);
        if (left != null) {
            if (
                (node.operator === "||" && Boolean(left.value) === true) ||
                (node.operator === "&&" && Boolean(left.value) === false) ||
                (node.operator === "??" && left.value != null)
            ) {
                return left
            ***REMOVED***

            const right = getStaticValueR(node.right, initialScope);
            if (right != null) {
                return right
            ***REMOVED***
        ***REMOVED***

        return null
    ***REMOVED***,

    MemberExpression(node, initialScope) {
        if (node.property.type === "PrivateIdentifier") {
            return null
        ***REMOVED***
        const object = getStaticValueR(node.object, initialScope);
        if (object != null) {
            if (object.value == null && (object.optional || node.optional)) {
                return { value: undefined, optional: true ***REMOVED***
            ***REMOVED***
            const property = getStaticPropertyNameValue(node, initialScope);

            if (property != null) {
                if (
                    !isGetter(
                        /** @type {object***REMOVED*** */ (object.value),
                        /** @type {PropertyKey***REMOVED*** */ (property.value),
                    )
                ) {
                    return {
                        value: /** @type {Record<PropertyKey, unknown>***REMOVED*** */ (
                            object.value
                        )[/** @type {PropertyKey***REMOVED*** */ (property.value)],
                    ***REMOVED***
                ***REMOVED***

                for (const [classFn, allowed] of getterAllowed) {
                    if (
                        object.value instanceof classFn &&
                        allowed.has(/** @type {string***REMOVED*** */ (property.value))
                    ) {
                        return {
                            value: /** @type {Record<PropertyKey, unknown>***REMOVED*** */ (
                                object.value
                            )[/** @type {PropertyKey***REMOVED*** */ (property.value)],
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        return null
    ***REMOVED***,

    ChainExpression(node, initialScope) {
        const expression = getStaticValueR(node.expression, initialScope);
        if (expression != null) {
            return { value: expression.value ***REMOVED***
        ***REMOVED***
        return null
    ***REMOVED***,

    NewExpression(node, initialScope) {
        const callee = getStaticValueR(node.callee, initialScope);
        const args = getElementValues(node.arguments, initialScope);

        if (callee != null && args != null) {
            const Func = /** @type {new (...args: any[]) => any***REMOVED*** */ (
                callee.value
            );
            if (callAllowed.has(Func)) {
                return { value: new Func(...args) ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        return null
    ***REMOVED***,

    ObjectExpression(node, initialScope) {
        /** @type {Record<PropertyKey, unknown>***REMOVED*** */
        const object = {***REMOVED***;

        for (const propertyNode of node.properties) {
            if (propertyNode.type === "Property") {
                if (propertyNode.kind !== "init") {
                    return null
                ***REMOVED***
                const key = getStaticPropertyNameValue(
                    propertyNode,
                    initialScope,
                );
                const value = getStaticValueR(propertyNode.value, initialScope);
                if (key == null || value == null) {
                    return null
                ***REMOVED***
                object[/** @type {PropertyKey***REMOVED*** */ (key.value)] = value.value;
            ***REMOVED*** else if (
                propertyNode.type === "SpreadElement" ||
                // @ts-expect-error -- Backward compatibility
                propertyNode.type === "ExperimentalSpreadProperty"
            ) {
                const argument = getStaticValueR(
                    propertyNode.argument,
                    initialScope,
                );
                if (argument == null) {
                    return null
                ***REMOVED***
                Object.assign(object, argument.value);
            ***REMOVED*** else {
                return null
            ***REMOVED***
        ***REMOVED***

        return { value: object ***REMOVED***
    ***REMOVED***,

    SequenceExpression(node, initialScope) {
        const last = node.expressions[node.expressions.length - 1];
        return getStaticValueR(last, initialScope)
    ***REMOVED***,

    TaggedTemplateExpression(node, initialScope) {
        const tag = getStaticValueR(node.tag, initialScope);
        const expressions = getElementValues(
            node.quasi.expressions,
            initialScope,
        );

        if (tag != null && expressions != null) {
            const func = /** @type {(...args: any[]) => any***REMOVED*** */ (tag.value);
            /** @type {any[] & { raw?: string[] ***REMOVED******REMOVED*** */
            const strings = node.quasi.quasis.map((q) => q.value.cooked);
            strings.raw = node.quasi.quasis.map((q) => q.value.raw);

            if (func === String.raw) {
                return { value: func(strings, ...expressions) ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        return null
    ***REMOVED***,

    TemplateLiteral(node, initialScope) {
        const expressions = getElementValues(node.expressions, initialScope);
        if (expressions != null) {
            let value = node.quasis[0].value.cooked;
            for (let i = 0; i < expressions.length; ++i) {
                value += expressions[i];
                value += /** @type {string***REMOVED*** */ (node.quasis[i + 1].value.cooked);
            ***REMOVED***
            return { value ***REMOVED***
        ***REMOVED***
        return null
    ***REMOVED***,

    UnaryExpression(node, initialScope) {
        if (node.operator === "delete") {
            // Not supported.
            return null
        ***REMOVED***
        if (node.operator === "void") {
            return { value: undefined ***REMOVED***
        ***REMOVED***

        const arg = getStaticValueR(node.argument, initialScope);
        if (arg != null) {
            switch (node.operator) {
                case "-":
                    return { value: -(/** @type {any***REMOVED*** */ (arg.value)) ***REMOVED***
                case "+":
                    return { value: +(/** @type {any***REMOVED*** */ (arg.value)) ***REMOVED*** //eslint-disable-line no-implicit-coercion
                case "!":
                    return { value: !arg.value ***REMOVED***
                case "~":
                    return { value: ~(/** @type {any***REMOVED*** */ (arg.value)) ***REMOVED***
                case "typeof":
                    return { value: typeof arg.value ***REMOVED***

                // no default
            ***REMOVED***
        ***REMOVED***

        return null
    ***REMOVED***,
    TSAsExpression(node, initialScope) {
        return getStaticValueR(node.expression, initialScope)
    ***REMOVED***,
    TSSatisfiesExpression(node, initialScope) {
        return getStaticValueR(node.expression, initialScope)
    ***REMOVED***,
    TSTypeAssertion(node, initialScope) {
        return getStaticValueR(node.expression, initialScope)
    ***REMOVED***,
    TSNonNullExpression(node, initialScope) {
        return getStaticValueR(node.expression, initialScope)
    ***REMOVED***,
    TSInstantiationExpression(node, initialScope) {
        return getStaticValueR(node.expression, initialScope)
    ***REMOVED***,
***REMOVED***);

/**
 * Get the value of a given node if it's a static value.
 * @param {Node|TSESTreeNode|null|undefined***REMOVED*** node The node to get.
 * @param {Scope|undefined|null***REMOVED*** initialScope The scope to start finding variable.
 * @returns {StaticValue|null***REMOVED*** The static value of the node, or `null`.
 */
function getStaticValueR(node, initialScope) {
    if (node != null && Object.hasOwnProperty.call(operations, node.type)) {
        return /** @type {VisitorCallback<any>***REMOVED*** */ (operations[node.type])(
            /** @type {TSESTreeNode***REMOVED*** */ (node),
            initialScope,
        )
    ***REMOVED***
    return null
***REMOVED***

/**
 * Get the static value of property name from a MemberExpression node or a Property node.
 * @param {MemberExpression|Property***REMOVED*** node The node to get.
 * @param {Scope|null***REMOVED*** [initialScope] The scope to start finding variable. Optional. If the node is a computed property node and this scope was given, this checks the computed property name by the `getStringIfConstant` function with the scope, and returns the value of it.
 * @returns {StaticValue|null***REMOVED*** The static value of the property name of the node, or `null`.
 */
function getStaticPropertyNameValue(node, initialScope) {
    const nameNode = node.type === "Property" ? node.key : node.property;

    if (node.computed) {
        return getStaticValueR(nameNode, initialScope)
    ***REMOVED***

    if (nameNode.type === "Identifier") {
        return { value: nameNode.name ***REMOVED***
    ***REMOVED***

    if (nameNode.type === "Literal") {
        if (/** @type {Partial<BigIntLiteral>***REMOVED*** */ (nameNode).bigint) {
            return { value: /** @type {BigIntLiteral***REMOVED*** */ (nameNode).bigint ***REMOVED***
        ***REMOVED***
        return { value: String(nameNode.value) ***REMOVED***
    ***REMOVED***

    return null
***REMOVED***

/**
 * Get the value of a given node if it's a static value.
 * @param {Node***REMOVED*** node The node to get.
 * @param {Scope|null***REMOVED*** [initialScope] The scope to start finding variable. Optional. If this scope was given, this tries to resolve identifier references which are in the given node as much as possible.
 * @returns {StaticValue | null***REMOVED*** The static value of the node, or `null`.
 */
function getStaticValue(node, initialScope = null) {
    try {
        return getStaticValueR(node, initialScope)
    ***REMOVED*** catch (_error) {
        return null
    ***REMOVED***
***REMOVED***

/** @typedef {import("eslint").Scope.Scope***REMOVED*** Scope */
/** @typedef {import("estree").Node***REMOVED*** Node */
/** @typedef {import("estree").RegExpLiteral***REMOVED*** RegExpLiteral */
/** @typedef {import("estree").BigIntLiteral***REMOVED*** BigIntLiteral */
/** @typedef {import("estree").SimpleLiteral***REMOVED*** SimpleLiteral */

/**
 * Get the value of a given node if it's a literal or a template literal.
 * @param {Node***REMOVED*** node The node to get.
 * @param {Scope|null***REMOVED*** [initialScope] The scope to start finding variable. Optional. If the node is an Identifier node and this scope was given, this checks the variable of the identifier, and returns the value of it if the variable is a constant.
 * @returns {string|null***REMOVED*** The value of the node, or `null`.
 */
function getStringIfConstant(node, initialScope = null) {
    // Handle the literals that the platform doesn't support natively.
    if (node && node.type === "Literal" && node.value === null) {
        const literal =
            /** @type {Partial<SimpleLiteral> & Partial<RegExpLiteral> & Partial<BigIntLiteral>***REMOVED*** */ (
                node
            );
        if (literal.regex) {
            return `/${literal.regex.pattern***REMOVED***/${literal.regex.flags***REMOVED***`
        ***REMOVED***
        if (literal.bigint) {
            return literal.bigint
        ***REMOVED***
    ***REMOVED***

    const evaluated = getStaticValue(node, initialScope);

    if (evaluated) {
        // `String(Symbol.prototype)` throws error
        try {
            return String(evaluated.value)
        ***REMOVED*** catch {
            // No op
        ***REMOVED***
    ***REMOVED***

    return null
***REMOVED***

/** @typedef {import("eslint").Scope.Scope***REMOVED*** Scope */
/** @typedef {import("estree").MemberExpression***REMOVED*** MemberExpression */
/** @typedef {import("estree").MethodDefinition***REMOVED*** MethodDefinition */
/** @typedef {import("estree").Property***REMOVED*** Property */
/** @typedef {import("estree").PropertyDefinition***REMOVED*** PropertyDefinition */
/** @typedef {import("estree").Identifier***REMOVED*** Identifier */

/**
 * Get the property name from a MemberExpression node or a Property node.
 * @param {MemberExpression | MethodDefinition | Property | PropertyDefinition***REMOVED*** node The node to get.
 * @param {Scope***REMOVED*** [initialScope] The scope to start finding variable. Optional. If the node is a computed property node and this scope was given, this checks the computed property name by the `getStringIfConstant` function with the scope, and returns the value of it.
 * @returns {string|null|undefined***REMOVED*** The property name of the node.
 */
function getPropertyName(node, initialScope) {
    switch (node.type) {
        case "MemberExpression":
            if (node.computed) {
                return getStringIfConstant(node.property, initialScope)
            ***REMOVED***
            if (node.property.type === "PrivateIdentifier") {
                return null
            ***REMOVED***
            return /** @type {Partial<Identifier>***REMOVED*** */ (node.property).name

        case "Property":
        case "MethodDefinition":
        case "PropertyDefinition":
            if (node.computed) {
                return getStringIfConstant(node.key, initialScope)
            ***REMOVED***
            if (node.key.type === "Literal") {
                return String(node.key.value)
            ***REMOVED***
            if (node.key.type === "PrivateIdentifier") {
                return null
            ***REMOVED***
            return /** @type {Partial<Identifier>***REMOVED*** */ (node.key).name
    ***REMOVED***

    return null
***REMOVED***

/** @typedef {import("eslint").Rule.Node***REMOVED*** RuleNode */
/** @typedef {import("eslint").SourceCode***REMOVED*** SourceCode */
/** @typedef {import("estree").Function***REMOVED*** FunctionNode */
/** @typedef {import("estree").FunctionDeclaration***REMOVED*** FunctionDeclaration */
/** @typedef {import("estree").FunctionExpression***REMOVED*** FunctionExpression */
/** @typedef {import("estree").Identifier***REMOVED*** Identifier */

/**
 * Get the name and kind of the given function node.
 * @param {FunctionNode***REMOVED*** node - The function node to get.
 * @param {SourceCode***REMOVED*** [sourceCode] The source code object to get the code of computed property keys.
 * @returns {string***REMOVED*** The name and kind of the function node.
 */
// eslint-disable-next-line complexity
function getFunctionNameWithKind(node, sourceCode) {
    const parent = /** @type {RuleNode***REMOVED*** */ (node).parent;
    const tokens = [];
    const isObjectMethod = parent.type === "Property" && parent.value === node;
    const isClassMethod =
        parent.type === "MethodDefinition" && parent.value === node;
    const isClassFieldMethod =
        parent.type === "PropertyDefinition" && parent.value === node;

    // Modifiers.
    if (isClassMethod || isClassFieldMethod) {
        if (parent.static) {
            tokens.push("static");
        ***REMOVED***
        if (parent.key.type === "PrivateIdentifier") {
            tokens.push("private");
        ***REMOVED***
    ***REMOVED***
    if (node.async) {
        tokens.push("async");
    ***REMOVED***
    if (node.generator) {
        tokens.push("generator");
    ***REMOVED***

    // Kinds.
    if (isObjectMethod || isClassMethod) {
        if (parent.kind === "constructor") {
            return "constructor"
        ***REMOVED***
        if (parent.kind === "get") {
            tokens.push("getter");
        ***REMOVED*** else if (parent.kind === "set") {
            tokens.push("setter");
        ***REMOVED*** else {
            tokens.push("method");
        ***REMOVED***
    ***REMOVED*** else if (isClassFieldMethod) {
        tokens.push("method");
    ***REMOVED*** else {
        if (node.type === "ArrowFunctionExpression") {
            tokens.push("arrow");
        ***REMOVED***
        tokens.push("function");
    ***REMOVED***

    // Names.
    if (isObjectMethod || isClassMethod || isClassFieldMethod) {
        if (parent.key.type === "PrivateIdentifier") {
            tokens.push(`#${parent.key.name***REMOVED***`);
        ***REMOVED*** else {
            const name = getPropertyName(parent);
            if (name) {
                tokens.push(`'${name***REMOVED***'`);
            ***REMOVED*** else if (sourceCode) {
                const keyText = sourceCode.getText(parent.key);
                if (!keyText.includes("\n")) {
                    tokens.push(`[${keyText***REMOVED***]`);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED*** else if (hasId(node)) {
        tokens.push(`'${node.id.name***REMOVED***'`);
    ***REMOVED*** else if (
        parent.type === "VariableDeclarator" &&
        parent.id &&
        parent.id.type === "Identifier"
    ) {
        tokens.push(`'${parent.id.name***REMOVED***'`);
    ***REMOVED*** else if (
        (parent.type === "AssignmentExpression" ||
            parent.type === "AssignmentPattern") &&
        parent.left &&
        parent.left.type === "Identifier"
    ) {
        tokens.push(`'${parent.left.name***REMOVED***'`);
    ***REMOVED*** else if (
        parent.type === "ExportDefaultDeclaration" &&
        parent.declaration === node
    ) {
        tokens.push("'default'");
    ***REMOVED***

    return tokens.join(" ")
***REMOVED***

/**
 * @param {FunctionNode***REMOVED*** node
 * @returns {node is FunctionDeclaration | FunctionExpression & { id: Identifier ***REMOVED******REMOVED***
 */
function hasId(node) {
    return Boolean(
        /** @type {Partial<FunctionDeclaration | FunctionExpression>***REMOVED*** */ (node)
            .id,
    )
***REMOVED***

/** @typedef {import("estree").Node***REMOVED*** Node */
/** @typedef {import("eslint").SourceCode***REMOVED*** SourceCode */
/** @typedef {import("./types.mjs").HasSideEffectOptions***REMOVED*** HasSideEffectOptions */
/** @typedef {import("estree").BinaryExpression***REMOVED*** BinaryExpression */
/** @typedef {import("estree").MemberExpression***REMOVED*** MemberExpression */
/** @typedef {import("estree").MethodDefinition***REMOVED*** MethodDefinition */
/** @typedef {import("estree").Property***REMOVED*** Property */
/** @typedef {import("estree").PropertyDefinition***REMOVED*** PropertyDefinition */
/** @typedef {import("estree").UnaryExpression***REMOVED*** UnaryExpression */

const typeConversionBinaryOps = Object.freeze(
    new Set([
        "==",
        "!=",
        "<",
        "<=",
        ">",
        ">=",
        "<<",
        ">>",
        ">>>",
        "+",
        "-",
        "*",
        "/",
        "%",
        "|",
        "^",
        "&",
        "in",
    ]),
);
const typeConversionUnaryOps = Object.freeze(new Set(["-", "+", "!", "~"]));

/**
 * Check whether the given value is an ASTNode or not.
 * @param {any***REMOVED*** x The value to check.
 * @returns {x is Node***REMOVED*** `true` if the value is an ASTNode.
 */
function isNode(x) {
    return x !== null && typeof x === "object" && typeof x.type === "string"
***REMOVED***

const visitor = Object.freeze(
    Object.assign(Object.create(null), {
        /**
         * @param {Node***REMOVED*** node
         * @param {HasSideEffectOptions***REMOVED*** options
         * @param {Record<string, string[]>***REMOVED*** visitorKeys
         */
        $visit(node, options, visitorKeys) {
            const { type ***REMOVED*** = node;

            if (typeof (/** @type {any***REMOVED*** */ (this)[type]) === "function") {
                return /** @type {any***REMOVED*** */ (this)[type](
                    node,
                    options,
                    visitorKeys,
                )
            ***REMOVED***

            return this.$visitChildren(node, options, visitorKeys)
        ***REMOVED***,

        /**
         * @param {Node***REMOVED*** node
         * @param {HasSideEffectOptions***REMOVED*** options
         * @param {Record<string, string[]>***REMOVED*** visitorKeys
         */
        $visitChildren(node, options, visitorKeys) {
            const { type ***REMOVED*** = node;

            for (const key of /** @type {(keyof Node)[]***REMOVED*** */ (
                visitorKeys[type] || eslintVisitorKeys.getKeys(node)
            )) {
                const value = node[key];

                if (Array.isArray(value)) {
                    for (const element of value) {
                        if (
                            isNode(element) &&
                            this.$visit(element, options, visitorKeys)
                        ) {
                            return true
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED*** else if (
                    isNode(value) &&
                    this.$visit(value, options, visitorKeys)
                ) {
                    return true
                ***REMOVED***
            ***REMOVED***

            return false
        ***REMOVED***,

        ArrowFunctionExpression() {
            return false
        ***REMOVED***,
        AssignmentExpression() {
            return true
        ***REMOVED***,
        AwaitExpression() {
            return true
        ***REMOVED***,
        /**
         * @param {BinaryExpression***REMOVED*** node
         * @param {HasSideEffectOptions***REMOVED*** options
         * @param {Record<string, string[]>***REMOVED*** visitorKeys
         */
        BinaryExpression(node, options, visitorKeys) {
            if (
                options.considerImplicitTypeConversion &&
                typeConversionBinaryOps.has(node.operator) &&
                (node.left.type !== "Literal" || node.right.type !== "Literal")
            ) {
                return true
            ***REMOVED***
            return this.$visitChildren(node, options, visitorKeys)
        ***REMOVED***,
        CallExpression() {
            return true
        ***REMOVED***,
        FunctionExpression() {
            return false
        ***REMOVED***,
        ImportExpression() {
            return true
        ***REMOVED***,
        /**
         * @param {MemberExpression***REMOVED*** node
         * @param {HasSideEffectOptions***REMOVED*** options
         * @param {Record<string, string[]>***REMOVED*** visitorKeys
         */
        MemberExpression(node, options, visitorKeys) {
            if (options.considerGetters) {
                return true
            ***REMOVED***
            if (
                options.considerImplicitTypeConversion &&
                node.computed &&
                node.property.type !== "Literal"
            ) {
                return true
            ***REMOVED***
            return this.$visitChildren(node, options, visitorKeys)
        ***REMOVED***,
        /**
         * @param {MethodDefinition***REMOVED*** node
         * @param {HasSideEffectOptions***REMOVED*** options
         * @param {Record<string, string[]>***REMOVED*** visitorKeys
         */
        MethodDefinition(node, options, visitorKeys) {
            if (
                options.considerImplicitTypeConversion &&
                node.computed &&
                node.key.type !== "Literal"
            ) {
                return true
            ***REMOVED***
            return this.$visitChildren(node, options, visitorKeys)
        ***REMOVED***,
        NewExpression() {
            return true
        ***REMOVED***,
        /**
         * @param {Property***REMOVED*** node
         * @param {HasSideEffectOptions***REMOVED*** options
         * @param {Record<string, string[]>***REMOVED*** visitorKeys
         */
        Property(node, options, visitorKeys) {
            if (
                options.considerImplicitTypeConversion &&
                node.computed &&
                node.key.type !== "Literal"
            ) {
                return true
            ***REMOVED***
            return this.$visitChildren(node, options, visitorKeys)
        ***REMOVED***,
        /**
         * @param {PropertyDefinition***REMOVED*** node
         * @param {HasSideEffectOptions***REMOVED*** options
         * @param {Record<string, string[]>***REMOVED*** visitorKeys
         */
        PropertyDefinition(node, options, visitorKeys) {
            if (
                options.considerImplicitTypeConversion &&
                node.computed &&
                node.key.type !== "Literal"
            ) {
                return true
            ***REMOVED***
            return this.$visitChildren(node, options, visitorKeys)
        ***REMOVED***,
        /**
         * @param {UnaryExpression***REMOVED*** node
         * @param {HasSideEffectOptions***REMOVED*** options
         * @param {Record<string, string[]>***REMOVED*** visitorKeys
         */
        UnaryExpression(node, options, visitorKeys) {
            if (node.operator === "delete") {
                return true
            ***REMOVED***
            if (
                options.considerImplicitTypeConversion &&
                typeConversionUnaryOps.has(node.operator) &&
                node.argument.type !== "Literal"
            ) {
                return true
            ***REMOVED***
            return this.$visitChildren(node, options, visitorKeys)
        ***REMOVED***,
        UpdateExpression() {
            return true
        ***REMOVED***,
        YieldExpression() {
            return true
        ***REMOVED***,
    ***REMOVED***),
);

/**
 * Check whether a given node has any side effect or not.
 * @param {Node***REMOVED*** node The node to get.
 * @param {SourceCode***REMOVED*** sourceCode The source code object.
 * @param {HasSideEffectOptions***REMOVED*** [options] The option object.
 * @returns {boolean***REMOVED*** `true` if the node has a certain side effect.
 */
function hasSideEffect(node, sourceCode, options = {***REMOVED***) {
    const { considerGetters = false, considerImplicitTypeConversion = false ***REMOVED*** =
        options;
    return visitor.$visit(
        node,
        { considerGetters, considerImplicitTypeConversion ***REMOVED***,
        sourceCode.visitorKeys || eslintVisitorKeys.KEYS,
    )
***REMOVED***

/** @typedef {import("estree").Node***REMOVED*** Node */
/** @typedef {import("eslint").SourceCode***REMOVED*** SourceCode */
/** @typedef {import("eslint").AST.Token***REMOVED*** Token */
/** @typedef {import("eslint").Rule.Node***REMOVED*** RuleNode */

/**
 * Get the left parenthesis of the parent node syntax if it exists.
 * E.g., `if (a) {***REMOVED***` then the `(`.
 * @param {Node***REMOVED*** node The AST node to check.
 * @param {SourceCode***REMOVED*** sourceCode The source code object to get tokens.
 * @returns {Token|null***REMOVED*** The left parenthesis of the parent node syntax
 */
function getParentSyntaxParen(node, sourceCode) {
    const parent = /** @type {RuleNode***REMOVED*** */ (node).parent;

    switch (parent.type) {
        case "CallExpression":
        case "NewExpression":
            if (parent.arguments.length === 1 && parent.arguments[0] === node) {
                return sourceCode.getTokenAfter(
                    parent.callee,
                    isOpeningParenToken,
                )
            ***REMOVED***
            return null

        case "DoWhileStatement":
            if (parent.test === node) {
                return sourceCode.getTokenAfter(
                    parent.body,
                    isOpeningParenToken,
                )
            ***REMOVED***
            return null

        case "IfStatement":
        case "WhileStatement":
            if (parent.test === node) {
                return sourceCode.getFirstToken(parent, 1)
            ***REMOVED***
            return null

        case "ImportExpression":
            if (parent.source === node) {
                return sourceCode.getFirstToken(parent, 1)
            ***REMOVED***
            return null

        case "SwitchStatement":
            if (parent.discriminant === node) {
                return sourceCode.getFirstToken(parent, 1)
            ***REMOVED***
            return null

        case "WithStatement":
            if (parent.object === node) {
                return sourceCode.getFirstToken(parent, 1)
            ***REMOVED***
            return null

        default:
            return null
    ***REMOVED***
***REMOVED***

/**
 * Check whether a given node is parenthesized or not.
 * @param {number***REMOVED*** times The number of parantheses.
 * @param {Node***REMOVED*** node The AST node to check.
 * @param {SourceCode***REMOVED*** sourceCode The source code object to get tokens.
 * @returns {boolean***REMOVED*** `true` if the node is parenthesized the given times.
 */
/**
 * Check whether a given node is parenthesized or not.
 * @param {Node***REMOVED*** node The AST node to check.
 * @param {SourceCode***REMOVED*** sourceCode The source code object to get tokens.
 * @returns {boolean***REMOVED*** `true` if the node is parenthesized.
 */
/**
 * Check whether a given node is parenthesized or not.
 * @param {Node|number***REMOVED*** timesOrNode The first parameter.
 * @param {Node|SourceCode***REMOVED*** nodeOrSourceCode The second parameter.
 * @param {SourceCode***REMOVED*** [optionalSourceCode] The third parameter.
 * @returns {boolean***REMOVED*** `true` if the node is parenthesized.
 */
function isParenthesized(
    timesOrNode,
    nodeOrSourceCode,
    optionalSourceCode,
) {
    /** @type {number***REMOVED*** */
    let times,
        /** @type {RuleNode***REMOVED*** */
        node,
        /** @type {SourceCode***REMOVED*** */
        sourceCode,
        maybeLeftParen,
        maybeRightParen;
    if (typeof timesOrNode === "number") {
        times = timesOrNode | 0;
        node = /** @type {RuleNode***REMOVED*** */ (nodeOrSourceCode);
        sourceCode = /** @type {SourceCode***REMOVED*** */ (optionalSourceCode);
        if (!(times >= 1)) {
            throw new TypeError("'times' should be a positive integer.")
        ***REMOVED***
    ***REMOVED*** else {
        times = 1;
        node = /** @type {RuleNode***REMOVED*** */ (timesOrNode);
        sourceCode = /** @type {SourceCode***REMOVED*** */ (nodeOrSourceCode);
    ***REMOVED***

    if (
        node == null ||
        // `Program` can't be parenthesized
        node.parent == null ||
        // `CatchClause.param` can't be parenthesized, example `try {***REMOVED*** catch (error) {***REMOVED***`
        (node.parent.type === "CatchClause" && node.parent.param === node)
    ) {
        return false
    ***REMOVED***

    maybeLeftParen = maybeRightParen = node;
    do {
        maybeLeftParen = sourceCode.getTokenBefore(maybeLeftParen);
        maybeRightParen = sourceCode.getTokenAfter(maybeRightParen);
    ***REMOVED*** while (
        maybeLeftParen != null &&
        maybeRightParen != null &&
        isOpeningParenToken(maybeLeftParen) &&
        isClosingParenToken(maybeRightParen) &&
        // Avoid false positive such as `if (a) {***REMOVED***`
        maybeLeftParen !== getParentSyntaxParen(node, sourceCode) &&
        --times > 0
    )

    return times === 0
***REMOVED***

/**
 * @author Toru Nagashima <https://github.com/mysticatea>
 * See LICENSE file in root directory for full license.
 */

const placeholder = /\$(?:[$&`']|[1-9][0-9]?)/gu;

/** @type {WeakMap<PatternMatcher, {pattern:RegExp,escaped:boolean***REMOVED***>***REMOVED*** */
const internal = new WeakMap();

/**
 * Check whether a given character is escaped or not.
 * @param {string***REMOVED*** str The string to check.
 * @param {number***REMOVED*** index The location of the character to check.
 * @returns {boolean***REMOVED*** `true` if the character is escaped.
 */
function isEscaped(str, index) {
    let escaped = false;
    for (let i = index - 1; i >= 0 && str.charCodeAt(i) === 0x5c; --i) {
        escaped = !escaped;
    ***REMOVED***
    return escaped
***REMOVED***

/**
 * Replace a given string by a given matcher.
 * @param {PatternMatcher***REMOVED*** matcher The pattern matcher.
 * @param {string***REMOVED*** str The string to be replaced.
 * @param {string***REMOVED*** replacement The new substring to replace each matched part.
 * @returns {string***REMOVED*** The replaced string.
 */
function replaceS(matcher, str, replacement) {
    const chunks = [];
    let index = 0;

    /**
     * @param {string***REMOVED*** key The placeholder.
     * @param {RegExpExecArray***REMOVED*** match The matched information.
     * @returns {string***REMOVED*** The replaced string.
     */
    function replacer(key, match) {
        switch (key) {
            case "$$":
                return "$"
            case "$&":
                return match[0]
            case "$`":
                return str.slice(0, match.index)
            case "$'":
                return str.slice(match.index + match[0].length)
            default: {
                const i = key.slice(1);
                if (i in match) {
                    return match[/** @type {any***REMOVED*** */ (i)]
                ***REMOVED***
                return key
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    for (const match of matcher.execAll(str)) {
        chunks.push(str.slice(index, match.index));
        chunks.push(
            replacement.replace(placeholder, (key) => replacer(key, match)),
        );
        index = match.index + match[0].length;
    ***REMOVED***
    chunks.push(str.slice(index));

    return chunks.join("")
***REMOVED***

/**
 * Replace a given string by a given matcher.
 * @param {PatternMatcher***REMOVED*** matcher The pattern matcher.
 * @param {string***REMOVED*** str The string to be replaced.
 * @param {(substring: string, ...args: any[]) => string***REMOVED*** replace The function to replace each matched part.
 * @returns {string***REMOVED*** The replaced string.
 */
function replaceF(matcher, str, replace) {
    const chunks = [];
    let index = 0;

    for (const match of matcher.execAll(str)) {
        chunks.push(str.slice(index, match.index));
        chunks.push(
            String(
                replace(
                    .../** @type {[string, ...string[]]***REMOVED*** */ (
                        /** @type {string[]***REMOVED*** */ (match)
                    ),
                    match.index,
                    match.input,
                ),
            ),
        );
        index = match.index + match[0].length;
    ***REMOVED***
    chunks.push(str.slice(index));

    return chunks.join("")
***REMOVED***

/**
 * The class to find patterns as considering escape sequences.
 */
class PatternMatcher {
    /**
     * Initialize this matcher.
     * @param {RegExp***REMOVED*** pattern The pattern to match.
     * @param {{escaped?:boolean***REMOVED******REMOVED*** [options] The options.
     */
    constructor(pattern, options = {***REMOVED***) {
        const { escaped = false ***REMOVED*** = options;
        if (!(pattern instanceof RegExp)) {
            throw new TypeError("'pattern' should be a RegExp instance.")
        ***REMOVED***
        if (!pattern.flags.includes("g")) {
            throw new Error("'pattern' should contains 'g' flag.")
        ***REMOVED***

        internal.set(this, {
            pattern: new RegExp(pattern.source, pattern.flags),
            escaped: Boolean(escaped),
        ***REMOVED***);
    ***REMOVED***

    /**
     * Find the pattern in a given string.
     * @param {string***REMOVED*** str The string to find.
     * @returns {IterableIterator<RegExpExecArray>***REMOVED*** The iterator which iterate the matched information.
     */
    *execAll(str) {
        const { pattern, escaped ***REMOVED*** =
            /** @type {{pattern:RegExp,escaped:boolean***REMOVED******REMOVED*** */ (internal.get(this));
        let match = null;
        let lastIndex = 0;

        pattern.lastIndex = 0;
        while ((match = pattern.exec(str)) != null) {
            if (escaped || !isEscaped(str, match.index)) {
                lastIndex = pattern.lastIndex;
                yield match;
                pattern.lastIndex = lastIndex;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    /**
     * Check whether the pattern is found in a given string.
     * @param {string***REMOVED*** str The string to check.
     * @returns {boolean***REMOVED*** `true` if the pattern was found in the string.
     */
    test(str) {
        const it = this.execAll(str);
        const ret = it.next();
        return !ret.done
    ***REMOVED***

    /**
     * Replace a given string.
     * @param {string***REMOVED*** str The string to be replaced.
     * @param {(string|((...strs:string[])=>string))***REMOVED*** replacer The string or function to replace. This is the same as the 2nd argument of `String.prototype.replace`.
     * @returns {string***REMOVED*** The replaced string.
     */
    [Symbol.replace](str, replacer) {
        return typeof replacer === "function"
            ? replaceF(this, String(str), replacer)
            : replaceS(this, String(str), String(replacer))
    ***REMOVED***
***REMOVED***

/** @typedef {import("eslint").Scope.Scope***REMOVED*** Scope */
/** @typedef {import("eslint").Scope.Variable***REMOVED*** Variable */
/** @typedef {import("eslint").Rule.Node***REMOVED*** RuleNode */
/** @typedef {import("estree").Node***REMOVED*** Node */
/** @typedef {import("estree").Expression***REMOVED*** Expression */
/** @typedef {import("estree").Pattern***REMOVED*** Pattern */
/** @typedef {import("estree").Identifier***REMOVED*** Identifier */
/** @typedef {import("estree").SimpleCallExpression***REMOVED*** CallExpression */
/** @typedef {import("estree").Program***REMOVED*** Program */
/** @typedef {import("estree").ImportDeclaration***REMOVED*** ImportDeclaration */
/** @typedef {import("estree").ExportAllDeclaration***REMOVED*** ExportAllDeclaration */
/** @typedef {import("estree").ExportDefaultDeclaration***REMOVED*** ExportDefaultDeclaration */
/** @typedef {import("estree").ExportNamedDeclaration***REMOVED*** ExportNamedDeclaration */
/** @typedef {import("estree").ImportSpecifier***REMOVED*** ImportSpecifier */
/** @typedef {import("estree").ImportDefaultSpecifier***REMOVED*** ImportDefaultSpecifier */
/** @typedef {import("estree").ImportNamespaceSpecifier***REMOVED*** ImportNamespaceSpecifier */
/** @typedef {import("estree").ExportSpecifier***REMOVED*** ExportSpecifier */
/** @typedef {import("estree").Property***REMOVED*** Property */
/** @typedef {import("estree").AssignmentProperty***REMOVED*** AssignmentProperty */
/** @typedef {import("estree").Literal***REMOVED*** Literal */
/** @typedef {import("@typescript-eslint/types").TSESTree.Node***REMOVED*** TSESTreeNode */
/** @typedef {import("./types.mjs").ReferenceTrackerOptions***REMOVED*** ReferenceTrackerOptions */
/**
 * @template T
 * @typedef {import("./types.mjs").TraceMap<T>***REMOVED*** TraceMap
 */
/**
 * @template T
 * @typedef {import("./types.mjs").TraceMapObject<T>***REMOVED*** TraceMapObject
 */
/**
 * @template T
 * @typedef {import("./types.mjs").TrackedReferences<T>***REMOVED*** TrackedReferences
 */

const IMPORT_TYPE = /^(?:Import|Export(?:All|Default|Named))Declaration$/u;

/**
 * Check whether a given node is an import node or not.
 * @param {Node***REMOVED*** node
 * @returns {node is ImportDeclaration|ExportAllDeclaration|ExportNamedDeclaration&{source: Literal***REMOVED******REMOVED*** `true` if the node is an import node.
 */
function isHasSource(node) {
    return (
        IMPORT_TYPE.test(node.type) &&
        /** @type {ImportDeclaration|ExportAllDeclaration|ExportNamedDeclaration***REMOVED*** */ (
            node
        ).source != null
    )
***REMOVED***
const has =
    /** @type {<T>(traceMap: TraceMap<unknown>, v: T) => v is (string extends T ? string : T)***REMOVED*** */ (
        Function.call.bind(Object.hasOwnProperty)
    );

const READ = Symbol("read");
const CALL = Symbol("call");
const CONSTRUCT = Symbol("construct");
const ESM = Symbol("esm");

const requireCall = { require: { [CALL]: true ***REMOVED*** ***REMOVED***;

/**
 * Check whether a given variable is modified or not.
 * @param {Variable|undefined***REMOVED*** variable The variable to check.
 * @returns {boolean***REMOVED*** `true` if the variable is modified.
 */
function isModifiedGlobal(variable) {
    return (
        variable == null ||
        variable.defs.length !== 0 ||
        variable.references.some((r) => r.isWrite())
    )
***REMOVED***

/**
 * Check if the value of a given node is passed through to the parent syntax as-is.
 * For example, `a` and `b` in (`a || b` and `c ? a : b`) are passed through.
 * @param {Node***REMOVED*** node A node to check.
 * @returns {node is RuleNode & {parent: Expression***REMOVED******REMOVED*** `true` if the node is passed through.
 */
function isPassThrough(node) {
    const parent = /** @type {TSESTreeNode***REMOVED*** */ (node).parent;

    if (parent) {
        switch (parent.type) {
            case "ConditionalExpression":
                return parent.consequent === node || parent.alternate === node
            case "LogicalExpression":
                return true
            case "SequenceExpression":
                return (
                    parent.expressions[parent.expressions.length - 1] === node
                )
            case "ChainExpression":
                return true
            case "TSAsExpression":
            case "TSSatisfiesExpression":
            case "TSTypeAssertion":
            case "TSNonNullExpression":
            case "TSInstantiationExpression":
                return true

            default:
                return false
        ***REMOVED***
    ***REMOVED***
    return false
***REMOVED***

/**
 * The reference tracker.
 */
class ReferenceTracker {
    /**
     * Initialize this tracker.
     * @param {Scope***REMOVED*** globalScope The global scope.
     * @param {object***REMOVED*** [options] The options.
     * @param {"legacy"|"strict"***REMOVED*** [options.mode="strict"] The mode to determine the ImportDeclaration's behavior for CJS modules.
     * @param {string[]***REMOVED*** [options.globalObjectNames=["global","globalThis","self","window"]] The variable names for Global Object.
     */
    constructor(globalScope, options = {***REMOVED***) {
        const {
            mode = "strict",
            globalObjectNames = ["global", "globalThis", "self", "window"],
        ***REMOVED*** = options;
        /** @private @type {Variable[]***REMOVED*** */
        this.variableStack = [];
        /** @private */
        this.globalScope = globalScope;
        /** @private */
        this.mode = mode;
        /** @private */
        this.globalObjectNames = globalObjectNames.slice(0);
    ***REMOVED***

    /**
     * Iterate the references of global variables.
     * @template T
     * @param {TraceMap<T>***REMOVED*** traceMap The trace map.
     * @returns {IterableIterator<TrackedReferences<T>>***REMOVED*** The iterator to iterate references.
     */
    *iterateGlobalReferences(traceMap) {
        for (const key of Object.keys(traceMap)) {
            const nextTraceMap = traceMap[key];
            const path = [key];
            const variable = this.globalScope.set.get(key);

            if (isModifiedGlobal(variable)) {
                continue
            ***REMOVED***

            yield* this._iterateVariableReferences(
                /** @type {Variable***REMOVED*** */ (variable),
                path,
                nextTraceMap,
                true,
            );
        ***REMOVED***

        for (const key of this.globalObjectNames) {
            /** @type {string[]***REMOVED*** */
            const path = [];
            const variable = this.globalScope.set.get(key);

            if (isModifiedGlobal(variable)) {
                continue
            ***REMOVED***

            yield* this._iterateVariableReferences(
                /** @type {Variable***REMOVED*** */ (variable),
                path,
                traceMap,
                false,
            );
        ***REMOVED***
    ***REMOVED***

    /**
     * Iterate the references of CommonJS modules.
     * @template T
     * @param {TraceMap<T>***REMOVED*** traceMap The trace map.
     * @returns {IterableIterator<TrackedReferences<T>>***REMOVED*** The iterator to iterate references.
     */
    *iterateCjsReferences(traceMap) {
        for (const { node ***REMOVED*** of this.iterateGlobalReferences(requireCall)) {
            const key = getStringIfConstant(
                /** @type {CallExpression***REMOVED*** */ (node).arguments[0],
            );
            if (key == null || !has(traceMap, key)) {
                continue
            ***REMOVED***

            const nextTraceMap = traceMap[key];
            const path = [key];

            if (nextTraceMap[READ]) {
                yield {
                    node,
                    path,
                    type: READ,
                    info: nextTraceMap[READ],
                ***REMOVED***;
            ***REMOVED***
            yield* this._iteratePropertyReferences(
                /** @type {CallExpression***REMOVED*** */ (node),
                path,
                nextTraceMap,
            );
        ***REMOVED***
    ***REMOVED***

    /**
     * Iterate the references of ES modules.
     * @template T
     * @param {TraceMap<T>***REMOVED*** traceMap The trace map.
     * @returns {IterableIterator<TrackedReferences<T>>***REMOVED*** The iterator to iterate references.
     */
    *iterateEsmReferences(traceMap) {
        const programNode = /** @type {Program***REMOVED*** */ (this.globalScope.block);

        for (const node of programNode.body) {
            if (!isHasSource(node)) {
                continue
            ***REMOVED***
            const moduleId = /** @type {string***REMOVED*** */ (node.source.value);

            if (!has(traceMap, moduleId)) {
                continue
            ***REMOVED***
            const nextTraceMap = traceMap[moduleId];
            const path = [moduleId];

            if (nextTraceMap[READ]) {
                yield {
                    // eslint-disable-next-line object-shorthand -- apply type
                    node: /** @type {RuleNode***REMOVED*** */ (node),
                    path,
                    type: READ,
                    info: nextTraceMap[READ],
                ***REMOVED***;
            ***REMOVED***

            if (node.type === "ExportAllDeclaration") {
                for (const key of Object.keys(nextTraceMap)) {
                    const exportTraceMap = nextTraceMap[key];
                    if (exportTraceMap[READ]) {
                        yield {
                            // eslint-disable-next-line object-shorthand -- apply type
                            node: /** @type {RuleNode***REMOVED*** */ (node),
                            path: path.concat(key),
                            type: READ,
                            info: exportTraceMap[READ],
                        ***REMOVED***;
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED*** else {
                for (const specifier of node.specifiers) {
                    const esm = has(nextTraceMap, ESM);
                    const it = this._iterateImportReferences(
                        specifier,
                        path,
                        esm
                            ? nextTraceMap
                            : this.mode === "legacy"
                            ? { default: nextTraceMap, ...nextTraceMap ***REMOVED***
                            : { default: nextTraceMap ***REMOVED***,
                    );

                    if (esm) {
                        yield* it;
                    ***REMOVED*** else {
                        for (const report of it) {
                            report.path = report.path.filter(exceptDefault);
                            if (
                                report.path.length >= 2 ||
                                report.type !== READ
                            ) {
                                yield report;
                            ***REMOVED***
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    /**
     * Iterate the property references for a given expression AST node.
     * @template T
     * @param {Expression***REMOVED*** node The expression AST node to iterate property references.
     * @param {TraceMap<T>***REMOVED*** traceMap The trace map.
     * @returns {IterableIterator<TrackedReferences<T>>***REMOVED*** The iterator to iterate property references.
     */
    *iteratePropertyReferences(node, traceMap) {
        yield* this._iteratePropertyReferences(node, [], traceMap);
    ***REMOVED***

    /**
     * Iterate the references for a given variable.
     * @private
     * @template T
     * @param {Variable***REMOVED*** variable The variable to iterate that references.
     * @param {string[]***REMOVED*** path The current path.
     * @param {TraceMapObject<T>***REMOVED*** traceMap The trace map.
     * @param {boolean***REMOVED*** shouldReport = The flag to report those references.
     * @returns {IterableIterator<TrackedReferences<T>>***REMOVED*** The iterator to iterate references.
     */
    *_iterateVariableReferences(variable, path, traceMap, shouldReport) {
        if (this.variableStack.includes(variable)) {
            return
        ***REMOVED***
        this.variableStack.push(variable);
        try {
            for (const reference of variable.references) {
                if (!reference.isRead()) {
                    continue
                ***REMOVED***
                const node = /** @type {RuleNode & Identifier***REMOVED*** */ (
                    reference.identifier
                );

                if (shouldReport && traceMap[READ]) {
                    yield { node, path, type: READ, info: traceMap[READ] ***REMOVED***;
                ***REMOVED***
                yield* this._iteratePropertyReferences(node, path, traceMap);
            ***REMOVED***
        ***REMOVED*** finally {
            this.variableStack.pop();
        ***REMOVED***
    ***REMOVED***

    /**
     * Iterate the references for a given AST node.
     * @private
     * @template T
     * @param {Expression***REMOVED*** rootNode The AST node to iterate references.
     * @param {string[]***REMOVED*** path The current path.
     * @param {TraceMapObject<T>***REMOVED*** traceMap The trace map.
     * @returns {IterableIterator<TrackedReferences<T>>***REMOVED*** The iterator to iterate references.
     */
    //eslint-disable-next-line complexity
    *_iteratePropertyReferences(rootNode, path, traceMap) {
        let node = rootNode;
        while (isPassThrough(node)) {
            node = node.parent;
        ***REMOVED***

        const parent = /** @type {RuleNode***REMOVED*** */ (node).parent;
        if (parent.type === "MemberExpression") {
            if (parent.object === node) {
                const key = getPropertyName(parent);
                if (key == null || !has(traceMap, key)) {
                    return
                ***REMOVED***

                path = path.concat(key); //eslint-disable-line no-param-reassign
                const nextTraceMap = traceMap[key];
                if (nextTraceMap[READ]) {
                    yield {
                        node: parent,
                        path,
                        type: READ,
                        info: nextTraceMap[READ],
                    ***REMOVED***;
                ***REMOVED***
                yield* this._iteratePropertyReferences(
                    parent,
                    path,
                    nextTraceMap,
                );
            ***REMOVED***
            return
        ***REMOVED***
        if (parent.type === "CallExpression") {
            if (parent.callee === node && traceMap[CALL]) {
                yield { node: parent, path, type: CALL, info: traceMap[CALL] ***REMOVED***;
            ***REMOVED***
            return
        ***REMOVED***
        if (parent.type === "NewExpression") {
            if (parent.callee === node && traceMap[CONSTRUCT]) {
                yield {
                    node: parent,
                    path,
                    type: CONSTRUCT,
                    info: traceMap[CONSTRUCT],
                ***REMOVED***;
            ***REMOVED***
            return
        ***REMOVED***
        if (parent.type === "AssignmentExpression") {
            if (parent.right === node) {
                yield* this._iterateLhsReferences(parent.left, path, traceMap);
                yield* this._iteratePropertyReferences(parent, path, traceMap);
            ***REMOVED***
            return
        ***REMOVED***
        if (parent.type === "AssignmentPattern") {
            if (parent.right === node) {
                yield* this._iterateLhsReferences(parent.left, path, traceMap);
            ***REMOVED***
            return
        ***REMOVED***
        if (parent.type === "VariableDeclarator") {
            if (parent.init === node) {
                yield* this._iterateLhsReferences(parent.id, path, traceMap);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    /**
     * Iterate the references for a given Pattern node.
     * @private
     * @template T
     * @param {Pattern***REMOVED*** patternNode The Pattern node to iterate references.
     * @param {string[]***REMOVED*** path The current path.
     * @param {TraceMapObject<T>***REMOVED*** traceMap The trace map.
     * @returns {IterableIterator<TrackedReferences<T>>***REMOVED*** The iterator to iterate references.
     */
    *_iterateLhsReferences(patternNode, path, traceMap) {
        if (patternNode.type === "Identifier") {
            const variable = findVariable(this.globalScope, patternNode);
            if (variable != null) {
                yield* this._iterateVariableReferences(
                    variable,
                    path,
                    traceMap,
                    false,
                );
            ***REMOVED***
            return
        ***REMOVED***
        if (patternNode.type === "ObjectPattern") {
            for (const property of patternNode.properties) {
                const key = getPropertyName(
                    /** @type {AssignmentProperty***REMOVED*** */ (property),
                );

                if (key == null || !has(traceMap, key)) {
                    continue
                ***REMOVED***

                const nextPath = path.concat(key);
                const nextTraceMap = traceMap[key];
                if (nextTraceMap[READ]) {
                    yield {
                        node: /** @type {RuleNode***REMOVED*** */ (property),
                        path: nextPath,
                        type: READ,
                        info: nextTraceMap[READ],
                    ***REMOVED***;
                ***REMOVED***
                yield* this._iterateLhsReferences(
                    /** @type {AssignmentProperty***REMOVED*** */ (property).value,
                    nextPath,
                    nextTraceMap,
                );
            ***REMOVED***
            return
        ***REMOVED***
        if (patternNode.type === "AssignmentPattern") {
            yield* this._iterateLhsReferences(patternNode.left, path, traceMap);
        ***REMOVED***
    ***REMOVED***

    /**
     * Iterate the references for a given ModuleSpecifier node.
     * @private
     * @template T
     * @param {ImportSpecifier | ImportDefaultSpecifier | ImportNamespaceSpecifier | ExportSpecifier***REMOVED*** specifierNode The ModuleSpecifier node to iterate references.
     * @param {string[]***REMOVED*** path The current path.
     * @param {TraceMapObject<T>***REMOVED*** traceMap The trace map.
     * @returns {IterableIterator<TrackedReferences<T>>***REMOVED*** The iterator to iterate references.
     */
    *_iterateImportReferences(specifierNode, path, traceMap) {
        const type = specifierNode.type;

        if (type === "ImportSpecifier" || type === "ImportDefaultSpecifier") {
            const key =
                type === "ImportDefaultSpecifier"
                    ? "default"
                    : specifierNode.imported.type === "Identifier"
                    ? specifierNode.imported.name
                    : specifierNode.imported.value;
            if (!has(traceMap, key)) {
                return
            ***REMOVED***

            path = path.concat(key); //eslint-disable-line no-param-reassign
            const nextTraceMap = traceMap[key];
            if (nextTraceMap[READ]) {
                yield {
                    node: /** @type {RuleNode***REMOVED*** */ (specifierNode),
                    path,
                    type: READ,
                    info: nextTraceMap[READ],
                ***REMOVED***;
            ***REMOVED***
            yield* this._iterateVariableReferences(
                /** @type {Variable***REMOVED*** */ (
                    findVariable(this.globalScope, specifierNode.local)
                ),
                path,
                nextTraceMap,
                false,
            );

            return
        ***REMOVED***

        if (type === "ImportNamespaceSpecifier") {
            yield* this._iterateVariableReferences(
                /** @type {Variable***REMOVED*** */ (
                    findVariable(this.globalScope, specifierNode.local)
                ),
                path,
                traceMap,
                false,
            );
            return
        ***REMOVED***

        if (type === "ExportSpecifier") {
            const key =
                specifierNode.local.type === "Identifier"
                    ? specifierNode.local.name
                    : specifierNode.local.value;
            if (!has(traceMap, key)) {
                return
            ***REMOVED***

            path = path.concat(key); //eslint-disable-line no-param-reassign
            const nextTraceMap = traceMap[key];
            if (nextTraceMap[READ]) {
                yield {
                    node: /** @type {RuleNode***REMOVED*** */ (specifierNode),
                    path,
                    type: READ,
                    info: nextTraceMap[READ],
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
***REMOVED***

ReferenceTracker.READ = READ;
ReferenceTracker.CALL = CALL;
ReferenceTracker.CONSTRUCT = CONSTRUCT;
ReferenceTracker.ESM = ESM;

/**
 * This is a predicate function for Array#filter.
 * @param {string***REMOVED*** name A name part.
 * @param {number***REMOVED*** index The index of the name.
 * @returns {boolean***REMOVED*** `false` if it's default.
 */
function exceptDefault(name, index) {
    return !(index === 1 && name === "default")
***REMOVED***

/** @typedef {import("./types.mjs").StaticValue***REMOVED*** StaticValue */

var index = {
    CALL,
    CONSTRUCT,
    ESM,
    findVariable,
    getFunctionHeadLocation,
    getFunctionNameWithKind,
    getInnermostScope,
    getPropertyName,
    getStaticValue,
    getStringIfConstant,
    hasSideEffect,
    isArrowToken,
    isClosingBraceToken,
    isClosingBracketToken,
    isClosingParenToken,
    isColonToken,
    isCommaToken,
    isCommentToken,
    isNotArrowToken,
    isNotClosingBraceToken,
    isNotClosingBracketToken,
    isNotClosingParenToken,
    isNotColonToken,
    isNotCommaToken,
    isNotCommentToken,
    isNotOpeningBraceToken,
    isNotOpeningBracketToken,
    isNotOpeningParenToken,
    isNotSemicolonToken,
    isOpeningBraceToken,
    isOpeningBracketToken,
    isOpeningParenToken,
    isParenthesized,
    isSemicolonToken,
    PatternMatcher,
    READ,
    ReferenceTracker,
***REMOVED***;

exports.CALL = CALL;
exports.CONSTRUCT = CONSTRUCT;
exports.ESM = ESM;
exports.PatternMatcher = PatternMatcher;
exports.READ = READ;
exports.ReferenceTracker = ReferenceTracker;
exports["default"] = index;
exports.findVariable = findVariable;
exports.getFunctionHeadLocation = getFunctionHeadLocation;
exports.getFunctionNameWithKind = getFunctionNameWithKind;
exports.getInnermostScope = getInnermostScope;
exports.getPropertyName = getPropertyName;
exports.getStaticValue = getStaticValue;
exports.getStringIfConstant = getStringIfConstant;
exports.hasSideEffect = hasSideEffect;
exports.isArrowToken = isArrowToken;
exports.isClosingBraceToken = isClosingBraceToken;
exports.isClosingBracketToken = isClosingBracketToken;
exports.isClosingParenToken = isClosingParenToken;
exports.isColonToken = isColonToken;
exports.isCommaToken = isCommaToken;
exports.isCommentToken = isCommentToken;
exports.isNotArrowToken = isNotArrowToken;
exports.isNotClosingBraceToken = isNotClosingBraceToken;
exports.isNotClosingBracketToken = isNotClosingBracketToken;
exports.isNotClosingParenToken = isNotClosingParenToken;
exports.isNotColonToken = isNotColonToken;
exports.isNotCommaToken = isNotCommaToken;
exports.isNotCommentToken = isNotCommentToken;
exports.isNotOpeningBraceToken = isNotOpeningBraceToken;
exports.isNotOpeningBracketToken = isNotOpeningBracketToken;
exports.isNotOpeningParenToken = isNotOpeningParenToken;
exports.isNotSemicolonToken = isNotSemicolonToken;
exports.isOpeningBraceToken = isOpeningBraceToken;
exports.isOpeningBracketToken = isOpeningBracketToken;
exports.isOpeningParenToken = isOpeningParenToken;
exports.isParenthesized = isParenthesized;
exports.isSemicolonToken = isSemicolonToken;
