// @babel/core/lib/config/files/plugins.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.loadPlugin = loadPlugin;
exports.loadPreset = loadPreset;
exports.resolvePreset = exports.resolvePlugin = void 0;
function _debug() {
  const data = require("debug");
  _debug = function () {
    return data;
  ***REMOVED***;
  return data;
***REMOVED***
function _path() {
  const data = require("path");
  _path = function () {
    return data;
  ***REMOVED***;
  return data;
***REMOVED***
var _async = require("../../gensync-utils/async.js");
var _moduleTypes = require("./module-types.js");
function _url() {
  const data = require("url");
  _url = function () {
    return data;
  ***REMOVED***;
  return data;
***REMOVED***
var _importMetaResolve = require("../../vendor/import-meta-resolve.js");
require("module");
function _fs() {
  const data = require("fs");
  _fs = function () {
    return data;
  ***REMOVED***;
  return data;
***REMOVED***
const debug = _debug()("babel:config:loading:files:plugins");
const EXACT_RE = /^module:/;
const BABEL_PLUGIN_PREFIX_RE = /^(?!@|module:|[^/]+\/|babel-plugin-)/;
const BABEL_PRESET_PREFIX_RE = /^(?!@|module:|[^/]+\/|babel-preset-)/;
const BABEL_PLUGIN_ORG_RE = /^(@babel\/)(?!plugin-|[^/]+\/)/;
const BABEL_PRESET_ORG_RE = /^(@babel\/)(?!preset-|[^/]+\/)/;
const OTHER_PLUGIN_ORG_RE = /^(@(?!babel\/)[^/]+\/)(?![^/]*babel-plugin(?:-|\/|$)|[^/]+\/)/;
const OTHER_PRESET_ORG_RE = /^(@(?!babel\/)[^/]+\/)(?![^/]*babel-preset(?:-|\/|$)|[^/]+\/)/;
const OTHER_ORG_DEFAULT_RE = /^(@(?!babel$)[^/]+)$/;
const resolvePlugin = exports.resolvePlugin = resolveStandardizedName.bind(null, "plugin");
const resolvePreset = exports.resolvePreset = resolveStandardizedName.bind(null, "preset");
function* loadPlugin(name, dirname) {
  const {
    filepath,
    loader
  ***REMOVED*** = resolvePlugin(name, dirname, yield* (0, _async.isAsync)());
  const value = yield* requireModule("plugin", loader, filepath);
  debug("Loaded plugin %o from %o.", name, dirname);
  return {
    filepath,
    value
  ***REMOVED***;
***REMOVED***
function* loadPreset(name, dirname) {
  const {
    filepath,
    loader
  ***REMOVED*** = resolvePreset(name, dirname, yield* (0, _async.isAsync)());
  const value = yield* requireModule("preset", loader, filepath);
  debug("Loaded preset %o from %o.", name, dirname);
  return {
    filepath,
    value
  ***REMOVED***;
***REMOVED***
function standardizeName(type, name) {
  if (_path().isAbsolute(name)) return name;
  const isPreset = type === "preset";
  return name.replace(isPreset ? BABEL_PRESET_PREFIX_RE : BABEL_PLUGIN_PREFIX_RE, `babel-${type***REMOVED***-`).replace(isPreset ? BABEL_PRESET_ORG_RE : BABEL_PLUGIN_ORG_RE, `$1${type***REMOVED***-`).replace(isPreset ? OTHER_PRESET_ORG_RE : OTHER_PLUGIN_ORG_RE, `$1babel-${type***REMOVED***-`).replace(OTHER_ORG_DEFAULT_RE, `$1/babel-${type***REMOVED***`).replace(EXACT_RE, "");
***REMOVED***
function* resolveAlternativesHelper(type, name) {
  const standardizedName = standardizeName(type, name);
  const {
    error,
    value
  ***REMOVED*** = yield standardizedName;
  if (!error) return value;
  if (error.code !== "MODULE_NOT_FOUND") throw error;
  if (standardizedName !== name && !(yield name).error) {
    error.message += `\n- If you want to resolve "${name***REMOVED***", use "module:${name***REMOVED***"`;
  ***REMOVED***
  if (!(yield standardizeName(type, "@babel/" + name)).error) {
    error.message += `\n- Did you mean "@babel/${name***REMOVED***"?`;
  ***REMOVED***
  const oppositeType = type === "preset" ? "plugin" : "preset";
  if (!(yield standardizeName(oppositeType, name)).error) {
    error.message += `\n- Did you accidentally pass a ${oppositeType***REMOVED*** as a ${type***REMOVED***?`;
  ***REMOVED***
  if (type === "plugin") {
    const transformName = standardizedName.replace("-proposal-", "-transform-");
    if (transformName !== standardizedName && !(yield transformName).error) {
      error.message += `\n- Did you mean "${transformName***REMOVED***"?`;
    ***REMOVED***
  ***REMOVED***
  error.message += `\n
Make sure that all the Babel plugins and presets you are using
are defined as dependencies or devDependencies in your package.json
file. It's possible that the missing plugin is loaded by a preset
you are using that forgot to add the plugin to its dependencies: you
can workaround this problem by explicitly adding the missing package
to your top-level package.json.
`;
  throw error;
***REMOVED***
function tryRequireResolve(id, dirname) {
  try {
    if (dirname) {
      return {
        error: null,
        value: (((v, w) => (v = v.split("."), w = w.split("."), +v[0] > +w[0] || v[0] == w[0] && +v[1] >= +w[1]))(process.versions.node, "8.9") ? require.resolve : (r, {
          paths: [b]
        ***REMOVED***, M = require("module")) => {
          let f = M._findPath(r, M._nodeModulePaths(b).concat(b));
          if (f) return f;
          f = new Error(`Cannot resolve module '${r***REMOVED***'`);
          f.code = "MODULE_NOT_FOUND";
          throw f;
        ***REMOVED***)(id, {
          paths: [dirname]
        ***REMOVED***)
      ***REMOVED***;
    ***REMOVED*** else {
      return {
        error: null,
        value: require.resolve(id)
      ***REMOVED***;
    ***REMOVED***
  ***REMOVED*** catch (error) {
    return {
      error,
      value: null
    ***REMOVED***;
  ***REMOVED***
***REMOVED***
function tryImportMetaResolve(id, options) {
  try {
    return {
      error: null,
      value: (0, _importMetaResolve.resolve)(id, options)
    ***REMOVED***;
  ***REMOVED*** catch (error) {
    return {
      error,
      value: null
    ***REMOVED***;
  ***REMOVED***
***REMOVED***
function resolveStandardizedNameForRequire(type, name, dirname) {
  const it = resolveAlternativesHelper(type, name);
  let res = it.next();
  while (!res.done) {
    res = it.next(tryRequireResolve(res.value, dirname));
  ***REMOVED***
  return {
    loader: "require",
    filepath: res.value
  ***REMOVED***;
***REMOVED***
function resolveStandardizedNameForImport(type, name, dirname) {
  const parentUrl = (0, _url().pathToFileURL)(_path().join(dirname, "./babel-virtual-resolve-base.js")).href;
  const it = resolveAlternativesHelper(type, name);
  let res = it.next();
  while (!res.done) {
    res = it.next(tryImportMetaResolve(res.value, parentUrl));
  ***REMOVED***
  return {
    loader: "auto",
    filepath: (0, _url().fileURLToPath)(res.value)
  ***REMOVED***;
***REMOVED***
function resolveStandardizedName(type, name, dirname, allowAsync) {
  if (!_moduleTypes.supportsESM || !allowAsync) {
    return resolveStandardizedNameForRequire(type, name, dirname);
  ***REMOVED***
  try {
    const resolved = resolveStandardizedNameForImport(type, name, dirname);
    if (!(0, _fs().existsSync)(resolved.filepath)) {
      throw Object.assign(new Error(`Could not resolve "${name***REMOVED***" in file ${dirname***REMOVED***.`), {
        type: "MODULE_NOT_FOUND"
      ***REMOVED***);
    ***REMOVED***
    return resolved;
  ***REMOVED*** catch (e) {
    try {
      return resolveStandardizedNameForRequire(type, name, dirname);
    ***REMOVED*** catch (e2) {
      if (e.type === "MODULE_NOT_FOUND") throw e;
      if (e2.type === "MODULE_NOT_FOUND") throw e2;
      throw e;
    ***REMOVED***
  ***REMOVED***
***REMOVED***
{
  var LOADING_MODULES = new Set();
***REMOVED***
function* requireModule(type, loader, name) {
  {
    if (!(yield* (0, _async.isAsync)()) && LOADING_MODULES.has(name)) {
      throw new Error(`Reentrant ${type***REMOVED*** detected trying to load "${name***REMOVED***". This module is not ignored ` + "and is trying to load itself while compiling itself, leading to a dependency cycle. " + 'We recommend adding it to your "ignore" list in your babelrc, or to a .babelignore.');
    ***REMOVED***
  ***REMOVED***
  try {
    {
      LOADING_MODULES.add(name);
    ***REMOVED***
    {
      return yield* (0, _moduleTypes.default)(name, loader, `You appear to be using a native ECMAScript module ${type***REMOVED***, ` + "which is only supported when running Babel asynchronously " + "or when using the Node.js `--experimental-require-module` flag.", `You appear to be using a ${type***REMOVED*** that contains top-level await, ` + "which is only supported when running Babel asynchronously.", true);
    ***REMOVED***
  ***REMOVED*** catch (err) {
    err.message = `[BABEL]: ${err.message***REMOVED*** (While processing: ${name***REMOVED***)`;
    throw err;
  ***REMOVED*** finally {
    {
      LOADING_MODULES.delete(name);
    ***REMOVED***
  ***REMOVED***
***REMOVED***
0 && 0;
