// ajv/dist/vocabularies/applicator/items.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.validateTuple = void 0;
const codegen_1 = require("../../compile/codegen");
const util_1 = require("../../compile/util");
const code_1 = require("../code");
const def = {
    keyword: "items",
    type: "array",
    schemaType: ["object", "array", "boolean"],
    before: "uniqueItems",
    code(cxt) {
        const { schema, it ***REMOVED*** = cxt;
        if (Array.isArray(schema))
            return validateTuple(cxt, "additionalItems", schema);
        it.items = true;
        if ((0, util_1.alwaysValidSchema)(it, schema))
            return;
        cxt.ok((0, code_1.validateArray)(cxt));
    ***REMOVED***,
***REMOVED***;
function validateTuple(cxt, extraItems, schArr = cxt.schema) {
    const { gen, parentSchema, data, keyword, it ***REMOVED*** = cxt;
    checkStrictTuple(parentSchema);
    if (it.opts.unevaluated && schArr.length && it.items !== true) {
        it.items = util_1.mergeEvaluated.items(gen, schArr.length, it.items);
    ***REMOVED***
    const valid = gen.name("valid");
    const len = gen.const("len", (0, codegen_1._) `${data***REMOVED***.length`);
    schArr.forEach((sch, i) => {
        if ((0, util_1.alwaysValidSchema)(it, sch))
            return;
        gen.if((0, codegen_1._) `${len***REMOVED*** > ${i***REMOVED***`, () => cxt.subschema({
            keyword,
            schemaProp: i,
            dataProp: i,
        ***REMOVED***, valid));
        cxt.ok(valid);
    ***REMOVED***);
    function checkStrictTuple(sch) {
        const { opts, errSchemaPath ***REMOVED*** = it;
        const l = schArr.length;
        const fullTuple = l === sch.minItems && (l === sch.maxItems || sch[extraItems] === false);
        if (opts.strictTuples && !fullTuple) {
            const msg = `"${keyword***REMOVED***" is ${l***REMOVED***-tuple, but minItems or maxItems/${extraItems***REMOVED*** are not specified or different at path "${errSchemaPath***REMOVED***"`;
            (0, util_1.checkStrictMode)(it, msg, opts.strictTuples);
        ***REMOVED***
    ***REMOVED***
***REMOVED***
exports.validateTuple = validateTuple;
exports.default = def;
