// ajv/dist/standalone/instance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const core_1 = require("../core");
const _1 = require(".");
const requireFromString = require("require-from-string");
class AjvPack {
    constructor(ajv) {
        this.ajv = ajv;
    ***REMOVED***
    validate(schemaKeyRef, data) {
        return core_1.default.prototype.validate.call(this, schemaKeyRef, data);
    ***REMOVED***
    compile(schema, meta) {
        return this.getStandalone(this.ajv.compile(schema, meta));
    ***REMOVED***
    getSchema(keyRef) {
        const v = this.ajv.getSchema(keyRef);
        if (!v)
            return undefined;
        return this.getStandalone(v);
    ***REMOVED***
    getStandalone(v) {
        return requireFromString((0, _1.default)(this.ajv, v));
    ***REMOVED***
    addSchema(...args) {
        this.ajv.addSchema.call(this.ajv, ...args);
        return this;
    ***REMOVED***
    addKeyword(...args) {
        this.ajv.addKeyword.call(this.ajv, ...args);
        return this;
    ***REMOVED***
***REMOVED***
exports.default = AjvPack;
