// webpack-sources/lib/OriginalSource.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

const { getMap, getSourceAndMap ***REMOVED*** = require("./helpers/getFromStreamChunks");
const splitIntoLines = require("./helpers/splitIntoLines");
const getGeneratedSourceInfo = require("./helpers/getGeneratedSourceInfo");
const Source = require("./Source");
const splitIntoPotentialTokens = require("./helpers/splitIntoPotentialTokens");

class OriginalSource extends Source {
	constructor(value, name) {
		super();
		const isBuffer = Buffer.isBuffer(value);
		this._value = isBuffer ? undefined : value;
		this._valueAsBuffer = isBuffer ? value : undefined;
		this._name = name;
	***REMOVED***

	getName() {
		return this._name;
	***REMOVED***

	source() {
		if (this._value === undefined) {
			this._value = this._valueAsBuffer.toString("utf-8");
		***REMOVED***
		return this._value;
	***REMOVED***

	buffer() {
		if (this._valueAsBuffer === undefined) {
			this._valueAsBuffer = Buffer.from(this._value, "utf-8");
		***REMOVED***
		return this._valueAsBuffer;
	***REMOVED***

	map(options) {
		return getMap(this, options);
	***REMOVED***

	sourceAndMap(options) {
		return getSourceAndMap(this, options);
	***REMOVED***

	/**
	 * @param {object***REMOVED*** options options
	 * @param {function(string, number, number, number, number, number, number): void***REMOVED*** onChunk called for each chunk of code
	 * @param {function(number, string, string)***REMOVED*** onSource called for each source
	 * @param {function(number, string)***REMOVED*** onName called for each name
	 * @returns {void***REMOVED***
	 */
	streamChunks(options, onChunk, onSource, onName) {
		if (this._value === undefined) {
			this._value = this._valueAsBuffer.toString("utf-8");
		***REMOVED***
		onSource(0, this._name, this._value);
		const finalSource = !!(options && options.finalSource);
		if (!options || options.columns !== false) {
			// With column info we need to read all lines and split them
			const matches = splitIntoPotentialTokens(this._value);
			let line = 1;
			let column = 0;
			if (matches !== null) {
				for (const match of matches) {
					const isEndOfLine = match.endsWith("\n");
					if (isEndOfLine && match.length === 1) {
						if (!finalSource) onChunk(match, line, column, -1, -1, -1, -1);
					***REMOVED*** else {
						const chunk = finalSource ? undefined : match;
						onChunk(chunk, line, column, 0, line, column, -1);
					***REMOVED***
					if (isEndOfLine) {
						line++;
						column = 0;
					***REMOVED*** else {
						column += match.length;
					***REMOVED***
				***REMOVED***
			***REMOVED***
			return {
				generatedLine: line,
				generatedColumn: column,
				source: finalSource ? this._value : undefined
			***REMOVED***;
		***REMOVED*** else if (finalSource) {
			// Without column info and with final source we only
			// need meta info to generate mapping
			const result = getGeneratedSourceInfo(this._value);
			const { generatedLine, generatedColumn ***REMOVED*** = result;
			if (generatedColumn === 0) {
				for (let line = 1; line < generatedLine; line++)
					onChunk(undefined, line, 0, 0, line, 0, -1);
			***REMOVED*** else {
				for (let line = 1; line <= generatedLine; line++)
					onChunk(undefined, line, 0, 0, line, 0, -1);
			***REMOVED***
			return result;
		***REMOVED*** else {
			// Without column info, but also without final source
			// we need to split source by lines
			let line = 1;
			const matches = splitIntoLines(this._value);
			let match;
			for (match of matches) {
				onChunk(finalSource ? undefined : match, line, 0, 0, line, 0, -1);
				line++;
			***REMOVED***
			return matches.length === 0 || match.endsWith("\n")
				? {
						generatedLine: matches.length + 1,
						generatedColumn: 0,
						source: finalSource ? this._value : undefined
				  ***REMOVED***
				: {
						generatedLine: matches.length,
						generatedColumn: match.length,
						source: finalSource ? this._value : undefined
				  ***REMOVED***;
		***REMOVED***
	***REMOVED***

	updateHash(hash) {
		if (this._valueAsBuffer === undefined) {
			this._valueAsBuffer = Buffer.from(this._value, "utf-8");
		***REMOVED***
		hash.update("OriginalSource");
		hash.update(this._valueAsBuffer);
		hash.update(this._name || "");
	***REMOVED***
***REMOVED***

module.exports = OriginalSource;
