// supports-hyperlinks/index.js
'use strict';
const supportsColor = require('supports-color');
const hasFlag = require('has-flag');

function parseVersion(versionString) {
	if (/^\d{3,4***REMOVED***$/.test(versionString)) {
		// Env var doesn't always use dots. example: 4601 => 46.1.0
		const m = /(\d{1,2***REMOVED***)(\d{2***REMOVED***)/.exec(versionString);
		return {
			major: 0,
			minor: parseInt(m[1], 10),
			patch: parseInt(m[2], 10)
		***REMOVED***;
	***REMOVED***

	const versions = (versionString || '').split('.').map(n => parseInt(n, 10));
	return {
		major: versions[0],
		minor: versions[1],
		patch: versions[2]
	***REMOVED***;
***REMOVED***

function supportsHyperlink(stream) {
	const {env***REMOVED*** = process;

	if ('FORCE_HYPERLINK' in env) {
		return !(env.FORCE_HYPERLINK.length > 0 && parseInt(env.FORCE_HYPERLINK, 10) === 0);
	***REMOVED***

	if (hasFlag('no-hyperlink') || hasFlag('no-hyperlinks') || hasFlag('hyperlink=false') || hasFlag('hyperlink=never')) {
		return false;
	***REMOVED***

	if (hasFlag('hyperlink=true') || hasFlag('hyperlink=always')) {
		return true;
	***REMOVED***

	// Netlify does not run a TTY, it does not need `supportsColor` check
	if ('NETLIFY' in env) {
		return true;
	***REMOVED***

	// If they specify no colors, they probably don't want hyperlinks.
	if (!supportsColor.supportsColor(stream)) {
		return false;
	***REMOVED***

	if (stream && !stream.isTTY) {
		return false;
	***REMOVED***

	if (process.platform === 'win32') {
		return false;
	***REMOVED***

	if ('CI' in env) {
		return false;
	***REMOVED***

	if ('TEAMCITY_VERSION' in env) {
		return false;
	***REMOVED***

	if ('TERM_PROGRAM' in env) {
		const version = parseVersion(env.TERM_PROGRAM_VERSION);

		switch (env.TERM_PROGRAM) {
			case 'iTerm.app':
				if (version.major === 3) {
					return version.minor >= 1;
				***REMOVED***

				return version.major > 3;
			case 'WezTerm':
				return version.major >= 20200620;
			case 'vscode':
				return version.major > 1 || version.major === 1 && version.minor >= 72;
			// No default
		***REMOVED***
	***REMOVED***

	if ('VTE_VERSION' in env) {
		// 0.50.0 was supposed to support hyperlinks, but throws a segfault
		if (env.VTE_VERSION === '0.50.0') {
			return false;
		***REMOVED***

		const version = parseVersion(env.VTE_VERSION);
		return version.major > 0 || version.minor >= 50;
	***REMOVED***

	return false;
***REMOVED***

module.exports = {
	supportsHyperlink,
	stdout: supportsHyperlink(process.stdout),
	stderr: supportsHyperlink(process.stderr)
***REMOVED***;
