// @svgr/babel-plugin-transform-react-native-svg/lib/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;
const elementToComponent = {
  svg: 'Svg',
  circle: 'Circle',
  clipPath: 'ClipPath',
  ellipse: 'Ellipse',
  g: 'G',
  linearGradient: 'LinearGradient',
  radialGradient: 'RadialGradient',
  line: 'Line',
  path: 'Path',
  pattern: 'Pattern',
  polygon: 'Polygon',
  polyline: 'Polyline',
  rect: 'Rect',
  symbol: 'Symbol',
  text: 'Text',
  textPath: 'TextPath',
  tspan: 'TSpan',
  use: 'Use',
  defs: 'Defs',
  stop: 'Stop',
  mask: 'Mask',
  image: 'Image',
  foreignObject: 'ForeignObject'
***REMOVED***;

const expoPrefix = (component, expo) => {
  // Prefix with 'Svg.' in the case we're transforming for Expo
  if (!expo) {
    return component;
  ***REMOVED***

  return (component !== 'Svg' ? 'Svg.' : '') + component;
***REMOVED***;

const plugin = ({
  types: t
***REMOVED***, {
  expo
***REMOVED***) => {
  function replaceElement(path, state) {
    const {
      name
    ***REMOVED*** = path.node.openingElement.name; // Replace element by react-native-svg components

    const component = elementToComponent[name];

    if (component) {
      const prefixedComponent = expoPrefix(component, expo);
      const openingElementName = path.get('openingElement.name');
      openingElementName.replaceWith(t.jsxIdentifier(prefixedComponent));

      if (path.has('closingElement')) {
        const closingElementName = path.get('closingElement.name');
        closingElementName.replaceWith(t.jsxIdentifier(prefixedComponent));
      ***REMOVED***

      state.replacedComponents.add(prefixedComponent);
      return;
    ***REMOVED*** // Remove element if not supported


    state.unsupportedComponents.add(name);
    path.remove();
  ***REMOVED***

  const svgElementVisitor = {
    JSXElement(path, state) {
      if (!path.get('openingElement.name').isJSXIdentifier({
        name: 'svg'
      ***REMOVED***)) {
        return;
      ***REMOVED***

      replaceElement(path, state);
      path.traverse(jsxElementVisitor, state);
    ***REMOVED***

  ***REMOVED***;
  const jsxElementVisitor = {
    JSXElement(path, state) {
      replaceElement(path, state);
    ***REMOVED***

  ***REMOVED***;
  const importDeclarationVisitor = {
    ImportDeclaration(path, state) {
      if (path.get('source').isStringLiteral({
        value: 'react-native-svg'
      ***REMOVED***)) {
        state.replacedComponents.forEach(component => {
          if (path.get('specifiers').some(specifier => specifier.get('local').isIdentifier({
            name: component
          ***REMOVED***))) {
            return;
          ***REMOVED***

          path.pushContainer('specifiers', t.importSpecifier(t.identifier(component), t.identifier(component)));
        ***REMOVED***);
      ***REMOVED*** else if (path.get('source').isStringLiteral({
        value: 'expo'
      ***REMOVED***)) {
        path.pushContainer('specifiers', t.importSpecifier(t.identifier('Svg'), t.identifier('Svg')));
      ***REMOVED*** else {
        return;
      ***REMOVED***

      if (state.unsupportedComponents.size && !path.has('trailingComments')) {
        const componentList = [...state.unsupportedComponents].join(', ');
        path.addComment('trailing', ` SVGR has dropped some elements not supported by react-native-svg: ${componentList***REMOVED*** `);
      ***REMOVED***
    ***REMOVED***

  ***REMOVED***;
  return {
    visitor: {
      Program(path, state) {
        state.replacedComponents = new Set();
        state.unsupportedComponents = new Set();
        path.traverse(svgElementVisitor, state);
        path.traverse(importDeclarationVisitor, state);
      ***REMOVED***

    ***REMOVED***
  ***REMOVED***;
***REMOVED***;

var _default = plugin;
exports.default = _default;
