// @eslint/eslintrc/lib/cascading-config-array-factory.js
/**
 * @fileoverview `CascadingConfigArrayFactory` class.
 *
 * `CascadingConfigArrayFactory` class has a responsibility:
 *
 * 1. Handles cascading of config files.
 *
 * It provides two methods:
 *
 * - `getConfigArrayForFile(filePath)`
 *     Get the corresponded configuration of a given file. This method doesn't
 *     throw even if the given file didn't exist.
 * - `clearCache()`
 *     Clear the internal cache. You have to call this method when
 *     `additionalPluginPool` was updated if `baseConfig` or `cliConfig` depends
 *     on the additional plugins. (`CLIEngine#addPlugin()` method calls this.)
 *
 * @author Toru Nagashima <https://github.com/mysticatea>
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const os = require("os");
const path = require("path");
const ConfigValidator = require("./shared/config-validator");
const { emitDeprecationWarning ***REMOVED*** = require("./shared/deprecation-warnings");
const { ConfigArrayFactory ***REMOVED*** = require("./config-array-factory");
const { ConfigArray, ConfigDependency, IgnorePattern ***REMOVED*** = require("./config-array");
const debug = require("debug")("eslintrc:cascading-config-array-factory");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

// Define types for VSCode IntelliSense.
/** @typedef {import("./shared/types").ConfigData***REMOVED*** ConfigData */
/** @typedef {import("./shared/types").Parser***REMOVED*** Parser */
/** @typedef {import("./shared/types").Plugin***REMOVED*** Plugin */
/** @typedef {import("./shared/types").Rule***REMOVED*** Rule */
/** @typedef {ReturnType<ConfigArrayFactory["create"]>***REMOVED*** ConfigArray */

/**
 * @typedef {Object***REMOVED*** CascadingConfigArrayFactoryOptions
 * @property {Map<string,Plugin>***REMOVED*** [additionalPluginPool] The map for additional plugins.
 * @property {ConfigData***REMOVED*** [baseConfig] The config by `baseConfig` option.
 * @property {ConfigData***REMOVED*** [cliConfig] The config by CLI options (`--env`, `--global`, `--ignore-pattern`, `--parser`, `--parser-options`, `--plugin`, and `--rule`). CLI options overwrite the setting in config files.
 * @property {string***REMOVED*** [cwd] The base directory to start lookup.
 * @property {string***REMOVED*** [ignorePath] The path to the alternative file of `.eslintignore`.
 * @property {string[]***REMOVED*** [rulePaths] The value of `--rulesdir` option.
 * @property {string***REMOVED*** [specificConfigPath] The value of `--config` option.
 * @property {boolean***REMOVED*** [useEslintrc] if `false` then it doesn't load config files.
 * @property {Function***REMOVED*** loadRules The function to use to load rules.
 * @property {Map<string,Rule>***REMOVED*** builtInRules The rules that are built in to ESLint.
 * @property {Object***REMOVED*** [resolver=ModuleResolver] The module resolver object.
 * @property {string***REMOVED*** eslintAllPath The path to the definitions for eslint:all.
 * @property {string***REMOVED*** eslintRecommendedPath The path to the definitions for eslint:recommended.
 */

/**
 * @typedef {Object***REMOVED*** CascadingConfigArrayFactoryInternalSlots
 * @property {ConfigArray***REMOVED*** baseConfigArray The config array of `baseConfig` option.
 * @property {ConfigData***REMOVED*** baseConfigData The config data of `baseConfig` option. This is used to reset `baseConfigArray`.
 * @property {ConfigArray***REMOVED*** cliConfigArray The config array of CLI options.
 * @property {ConfigData***REMOVED*** cliConfigData The config data of CLI options. This is used to reset `cliConfigArray`.
 * @property {ConfigArrayFactory***REMOVED*** configArrayFactory The factory for config arrays.
 * @property {Map<string, ConfigArray>***REMOVED*** configCache The cache from directory paths to config arrays.
 * @property {string***REMOVED*** cwd The base directory to start lookup.
 * @property {WeakMap<ConfigArray, ConfigArray>***REMOVED*** finalizeCache The cache from config arrays to finalized config arrays.
 * @property {string***REMOVED*** [ignorePath] The path to the alternative file of `.eslintignore`.
 * @property {string[]|null***REMOVED*** rulePaths The value of `--rulesdir` option. This is used to reset `baseConfigArray`.
 * @property {string|null***REMOVED*** specificConfigPath The value of `--config` option. This is used to reset `cliConfigArray`.
 * @property {boolean***REMOVED*** useEslintrc if `false` then it doesn't load config files.
 * @property {Function***REMOVED*** loadRules The function to use to load rules.
 * @property {Map<string,Rule>***REMOVED*** builtInRules The rules that are built in to ESLint.
 * @property {Object***REMOVED*** [resolver=ModuleResolver] The module resolver object.
 * @property {string***REMOVED*** eslintAllPath The path to the definitions for eslint:all.
 * @property {string***REMOVED*** eslintRecommendedPath The path to the definitions for eslint:recommended.
 */

/** @type {WeakMap<CascadingConfigArrayFactory, CascadingConfigArrayFactoryInternalSlots>***REMOVED*** */
const internalSlotsMap = new WeakMap();

/**
 * Create the config array from `baseConfig` and `rulePaths`.
 * @param {CascadingConfigArrayFactoryInternalSlots***REMOVED*** slots The slots.
 * @returns {ConfigArray***REMOVED*** The config array of the base configs.
 */
function createBaseConfigArray({
    configArrayFactory,
    baseConfigData,
    rulePaths,
    cwd,
    loadRules
***REMOVED***) {
    const baseConfigArray = configArrayFactory.create(
        baseConfigData,
        { name: "BaseConfig" ***REMOVED***
    );

    /*
     * Create the config array element for the default ignore patterns.
     * This element has `ignorePattern` property that ignores the default
     * patterns in the current working directory.
     */
    baseConfigArray.unshift(configArrayFactory.create(
        { ignorePatterns: IgnorePattern.DefaultPatterns ***REMOVED***,
        { name: "DefaultIgnorePattern" ***REMOVED***
    )[0]);

    /*
     * Load rules `--rulesdir` option as a pseudo plugin.
     * Use a pseudo plugin to define rules of `--rulesdir`, so we can validate
     * the rule's options with only information in the config array.
     */
    if (rulePaths && rulePaths.length > 0) {
        baseConfigArray.push({
            type: "config",
            name: "--rulesdir",
            filePath: "",
            plugins: {
                "": new ConfigDependency({
                    definition: {
                        rules: rulePaths.reduce(
                            (map, rulesPath) => Object.assign(
                                map,
                                loadRules(rulesPath, cwd)
                            ),
                            {***REMOVED***
                        )
                    ***REMOVED***,
                    filePath: "",
                    id: "",
                    importerName: "--rulesdir",
                    importerPath: ""
                ***REMOVED***)
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***

    return baseConfigArray;
***REMOVED***

/**
 * Create the config array from CLI options.
 * @param {CascadingConfigArrayFactoryInternalSlots***REMOVED*** slots The slots.
 * @returns {ConfigArray***REMOVED*** The config array of the base configs.
 */
function createCLIConfigArray({
    cliConfigData,
    configArrayFactory,
    cwd,
    ignorePath,
    specificConfigPath
***REMOVED***) {
    const cliConfigArray = configArrayFactory.create(
        cliConfigData,
        { name: "CLIOptions" ***REMOVED***
    );

    cliConfigArray.unshift(
        ...(ignorePath
            ? configArrayFactory.loadESLintIgnore(ignorePath)
            : configArrayFactory.loadDefaultESLintIgnore())
    );

    if (specificConfigPath) {
        cliConfigArray.unshift(
            ...configArrayFactory.loadFile(
                specificConfigPath,
                { name: "--config", basePath: cwd ***REMOVED***
            )
        );
    ***REMOVED***

    return cliConfigArray;
***REMOVED***

/**
 * The error type when there are files matched by a glob, but all of them have been ignored.
 */
class ConfigurationNotFoundError extends Error {

    // eslint-disable-next-line jsdoc/require-description
    /**
     * @param {string***REMOVED*** directoryPath The directory path.
     */
    constructor(directoryPath) {
        super(`No ESLint configuration found in ${directoryPath***REMOVED***.`);
        this.messageTemplate = "no-config-found";
        this.messageData = { directoryPath ***REMOVED***;
    ***REMOVED***
***REMOVED***

/**
 * This class provides the functionality that enumerates every file which is
 * matched by given glob patterns and that configuration.
 */
class CascadingConfigArrayFactory {

    /**
     * Initialize this enumerator.
     * @param {CascadingConfigArrayFactoryOptions***REMOVED*** options The options.
     */
    constructor({
        additionalPluginPool = new Map(),
        baseConfig: baseConfigData = null,
        cliConfig: cliConfigData = null,
        cwd = process.cwd(),
        ignorePath,
        resolvePluginsRelativeTo,
        rulePaths = [],
        specificConfigPath = null,
        useEslintrc = true,
        builtInRules = new Map(),
        loadRules,
        resolver,
        eslintRecommendedPath,
        eslintAllPath
    ***REMOVED*** = {***REMOVED***) {
        const configArrayFactory = new ConfigArrayFactory({
            additionalPluginPool,
            cwd,
            resolvePluginsRelativeTo,
            builtInRules,
            resolver,
            eslintRecommendedPath,
            eslintAllPath
        ***REMOVED***);

        internalSlotsMap.set(this, {
            baseConfigArray: createBaseConfigArray({
                baseConfigData,
                configArrayFactory,
                cwd,
                rulePaths,
                loadRules,
                resolver
            ***REMOVED***),
            baseConfigData,
            cliConfigArray: createCLIConfigArray({
                cliConfigData,
                configArrayFactory,
                cwd,
                ignorePath,
                specificConfigPath
            ***REMOVED***),
            cliConfigData,
            configArrayFactory,
            configCache: new Map(),
            cwd,
            finalizeCache: new WeakMap(),
            ignorePath,
            rulePaths,
            specificConfigPath,
            useEslintrc,
            builtInRules,
            loadRules
        ***REMOVED***);
    ***REMOVED***

    /**
     * The path to the current working directory.
     * This is used by tests.
     * @type {string***REMOVED***
     */
    get cwd() {
        const { cwd ***REMOVED*** = internalSlotsMap.get(this);

        return cwd;
    ***REMOVED***

    /**
     * Get the config array of a given file.
     * If `filePath` was not given, it returns the config which contains only
     * `baseConfigData` and `cliConfigData`.
     * @param {string***REMOVED*** [filePath] The file path to a file.
     * @param {Object***REMOVED*** [options] The options.
     * @param {boolean***REMOVED*** [options.ignoreNotFoundError] If `true` then it doesn't throw `ConfigurationNotFoundError`.
     * @returns {ConfigArray***REMOVED*** The config array of the file.
     */
    getConfigArrayForFile(filePath, { ignoreNotFoundError = false ***REMOVED*** = {***REMOVED***) {
        const {
            baseConfigArray,
            cliConfigArray,
            cwd
        ***REMOVED*** = internalSlotsMap.get(this);

        if (!filePath) {
            return new ConfigArray(...baseConfigArray, ...cliConfigArray);
        ***REMOVED***

        const directoryPath = path.dirname(path.resolve(cwd, filePath));

        debug(`Load config files for ${directoryPath***REMOVED***.`);

        return this._finalizeConfigArray(
            this._loadConfigInAncestors(directoryPath),
            directoryPath,
            ignoreNotFoundError
        );
    ***REMOVED***

    /**
     * Set the config data to override all configs.
     * Require to call `clearCache()` method after this method is called.
     * @param {ConfigData***REMOVED*** configData The config data to override all configs.
     * @returns {void***REMOVED***
     */
    setOverrideConfig(configData) {
        const slots = internalSlotsMap.get(this);

        slots.cliConfigData = configData;
    ***REMOVED***

    /**
     * Clear config cache.
     * @returns {void***REMOVED***
     */
    clearCache() {
        const slots = internalSlotsMap.get(this);

        slots.baseConfigArray = createBaseConfigArray(slots);
        slots.cliConfigArray = createCLIConfigArray(slots);
        slots.configCache.clear();
    ***REMOVED***

    /**
     * Load and normalize config files from the ancestor directories.
     * @param {string***REMOVED*** directoryPath The path to a leaf directory.
     * @param {boolean***REMOVED*** configsExistInSubdirs `true` if configurations exist in subdirectories.
     * @returns {ConfigArray***REMOVED*** The loaded config.
     * @private
     */
    _loadConfigInAncestors(directoryPath, configsExistInSubdirs = false) {
        const {
            baseConfigArray,
            configArrayFactory,
            configCache,
            cwd,
            useEslintrc
        ***REMOVED*** = internalSlotsMap.get(this);

        if (!useEslintrc) {
            return baseConfigArray;
        ***REMOVED***

        let configArray = configCache.get(directoryPath);

        // Hit cache.
        if (configArray) {
            debug(`Cache hit: ${directoryPath***REMOVED***.`);
            return configArray;
        ***REMOVED***
        debug(`No cache found: ${directoryPath***REMOVED***.`);

        const homePath = os.homedir();

        // Consider this is root.
        if (directoryPath === homePath && cwd !== homePath) {
            debug("Stop traversing because of considered root.");
            if (configsExistInSubdirs) {
                const filePath = ConfigArrayFactory.getPathToConfigFileInDirectory(directoryPath);

                if (filePath) {
                    emitDeprecationWarning(
                        filePath,
                        "ESLINT_PERSONAL_CONFIG_SUPPRESS"
                    );
                ***REMOVED***
            ***REMOVED***
            return this._cacheConfig(directoryPath, baseConfigArray);
        ***REMOVED***

        // Load the config on this directory.
        try {
            configArray = configArrayFactory.loadInDirectory(directoryPath);
        ***REMOVED*** catch (error) {
            /* istanbul ignore next */
            if (error.code === "EACCES") {
                debug("Stop traversing because of 'EACCES' error.");
                return this._cacheConfig(directoryPath, baseConfigArray);
            ***REMOVED***
            throw error;
        ***REMOVED***

        if (configArray.length > 0 && configArray.isRoot()) {
            debug("Stop traversing because of 'root:true'.");
            configArray.unshift(...baseConfigArray);
            return this._cacheConfig(directoryPath, configArray);
        ***REMOVED***

        // Load from the ancestors and merge it.
        const parentPath = path.dirname(directoryPath);
        const parentConfigArray = parentPath && parentPath !== directoryPath
            ? this._loadConfigInAncestors(
                parentPath,
                configsExistInSubdirs || configArray.length > 0
            )
            : baseConfigArray;

        if (configArray.length > 0) {
            configArray.unshift(...parentConfigArray);
        ***REMOVED*** else {
            configArray = parentConfigArray;
        ***REMOVED***

        // Cache and return.
        return this._cacheConfig(directoryPath, configArray);
    ***REMOVED***

    /**
     * Freeze and cache a given config.
     * @param {string***REMOVED*** directoryPath The path to a directory as a cache key.
     * @param {ConfigArray***REMOVED*** configArray The config array as a cache value.
     * @returns {ConfigArray***REMOVED*** The `configArray` (frozen).
     */
    _cacheConfig(directoryPath, configArray) {
        const { configCache ***REMOVED*** = internalSlotsMap.get(this);

        Object.freeze(configArray);
        configCache.set(directoryPath, configArray);

        return configArray;
    ***REMOVED***

    /**
     * Finalize a given config array.
     * Concatenate `--config` and other CLI options.
     * @param {ConfigArray***REMOVED*** configArray The parent config array.
     * @param {string***REMOVED*** directoryPath The path to the leaf directory to find config files.
     * @param {boolean***REMOVED*** ignoreNotFoundError If `true` then it doesn't throw `ConfigurationNotFoundError`.
     * @returns {ConfigArray***REMOVED*** The loaded config.
     * @private
     */
    _finalizeConfigArray(configArray, directoryPath, ignoreNotFoundError) {
        const {
            cliConfigArray,
            configArrayFactory,
            finalizeCache,
            useEslintrc,
            builtInRules
        ***REMOVED*** = internalSlotsMap.get(this);

        let finalConfigArray = finalizeCache.get(configArray);

        if (!finalConfigArray) {
            finalConfigArray = configArray;

            // Load the personal config if there are no regular config files.
            if (
                useEslintrc &&
                configArray.every(c => !c.filePath) &&
                cliConfigArray.every(c => !c.filePath) // `--config` option can be a file.
            ) {
                const homePath = os.homedir();

                debug("Loading the config file of the home directory:", homePath);

                const personalConfigArray = configArrayFactory.loadInDirectory(
                    homePath,
                    { name: "PersonalConfig" ***REMOVED***
                );

                if (
                    personalConfigArray.length > 0 &&
                    !directoryPath.startsWith(homePath)
                ) {
                    const lastElement =
                        personalConfigArray[personalConfigArray.length - 1];

                    emitDeprecationWarning(
                        lastElement.filePath,
                        "ESLINT_PERSONAL_CONFIG_LOAD"
                    );
                ***REMOVED***

                finalConfigArray = finalConfigArray.concat(personalConfigArray);
            ***REMOVED***

            // Apply CLI options.
            if (cliConfigArray.length > 0) {
                finalConfigArray = finalConfigArray.concat(cliConfigArray);
            ***REMOVED***

            // Validate rule settings and environments.
            const validator = new ConfigValidator({
                builtInRules
            ***REMOVED***);

            validator.validateConfigArray(finalConfigArray);

            // Cache it.
            Object.freeze(finalConfigArray);
            finalizeCache.set(configArray, finalConfigArray);

            debug(
                "Configuration was determined: %o on %s",
                finalConfigArray,
                directoryPath
            );
        ***REMOVED***

        // At least one element (the default ignore patterns) exists.
        if (!ignoreNotFoundError && useEslintrc && finalConfigArray.length <= 1) {
            throw new ConfigurationNotFoundError(directoryPath);
        ***REMOVED***

        return finalConfigArray;
    ***REMOVED***
***REMOVED***

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

module.exports = { CascadingConfigArrayFactory ***REMOVED***;
