// @eslint/eslintrc/lib/config-array-factory.js
/**
 * @fileoverview The factory of `ConfigArray` objects.
 *
 * This class provides methods to create `ConfigArray` instance.
 *
 * - `create(configData, options)`
 *     Create a `ConfigArray` instance from a config data. This is to handle CLI
 *     options except `--config`.
 * - `loadFile(filePath, options)`
 *     Create a `ConfigArray` instance from a config file. This is to handle
 *     `--config` option. If the file was not found, throws the following error:
 *      - If the filename was `*.js`, a `MODULE_NOT_FOUND` error.
 *      - If the filename was `package.json`, an IO error or an
 *        `ESLINT_CONFIG_FIELD_NOT_FOUND` error.
 *      - Otherwise, an IO error such as `ENOENT`.
 * - `loadInDirectory(directoryPath, options)`
 *     Create a `ConfigArray` instance from a config file which is on a given
 *     directory. This tries to load `.eslintrc.*` or `package.json`. If not
 *     found, returns an empty `ConfigArray`.
 * - `loadESLintIgnore(filePath)`
 *     Create a `ConfigArray` instance from a config file that is `.eslintignore`
 *     format. This is to handle `--ignore-path` option.
 * - `loadDefaultESLintIgnore()`
 *     Create a `ConfigArray` instance from `.eslintignore` or `package.json` in
 *     the current working directory.
 *
 * `ConfigArrayFactory` class has the responsibility that loads configuration
 * files, including loading `extends`, `parser`, and `plugins`. The created
 * `ConfigArray` instance has the loaded `extends`, `parser`, and `plugins`.
 *
 * But this class doesn't handle cascading. `CascadingConfigArrayFactory` class
 * handles cascading and hierarchy.
 *
 * @author Toru Nagashima <https://github.com/mysticatea>
 */

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

import debugOrig from "debug";
import fs from "node:fs";
import importFresh from "import-fresh";
import { createRequire ***REMOVED*** from "node:module";
import path from "node:path";
import stripComments from "strip-json-comments";

import {
    ConfigArray,
    ConfigDependency,
    IgnorePattern,
    OverrideTester
***REMOVED*** from "./config-array/index.js";
import ConfigValidator from "./shared/config-validator.js";
import * as naming from "./shared/naming.js";
import * as ModuleResolver from "./shared/relative-module-resolver.js";

const require = createRequire(import.meta.url);

const debug = debugOrig("eslintrc:config-array-factory");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

const configFilenames = [
    ".eslintrc.js",
    ".eslintrc.cjs",
    ".eslintrc.yaml",
    ".eslintrc.yml",
    ".eslintrc.json",
    ".eslintrc",
    "package.json"
];

// Define types for VSCode IntelliSense.
/** @typedef {import("./shared/types").ConfigData***REMOVED*** ConfigData */
/** @typedef {import("./shared/types").OverrideConfigData***REMOVED*** OverrideConfigData */
/** @typedef {import("./shared/types").Parser***REMOVED*** Parser */
/** @typedef {import("./shared/types").Plugin***REMOVED*** Plugin */
/** @typedef {import("./shared/types").Rule***REMOVED*** Rule */
/** @typedef {import("./config-array/config-dependency").DependentParser***REMOVED*** DependentParser */
/** @typedef {import("./config-array/config-dependency").DependentPlugin***REMOVED*** DependentPlugin */
/** @typedef {ConfigArray[0]***REMOVED*** ConfigArrayElement */

/**
 * @typedef {Object***REMOVED*** ConfigArrayFactoryOptions
 * @property {Map<string,Plugin>***REMOVED*** [additionalPluginPool] The map for additional plugins.
 * @property {string***REMOVED*** [cwd] The path to the current working directory.
 * @property {string***REMOVED*** [resolvePluginsRelativeTo] A path to the directory that plugins should be resolved from. Defaults to `cwd`.
 * @property {Map<string,Rule>***REMOVED*** builtInRules The rules that are built in to ESLint.
 * @property {Object***REMOVED*** [resolver=ModuleResolver] The module resolver object.
 * @property {string***REMOVED*** eslintAllPath The path to the definitions for eslint:all.
 * @property {Function***REMOVED*** getEslintAllConfig Returns the config data for eslint:all.
 * @property {string***REMOVED*** eslintRecommendedPath The path to the definitions for eslint:recommended.
 * @property {Function***REMOVED*** getEslintRecommendedConfig Returns the config data for eslint:recommended.
 */

/**
 * @typedef {Object***REMOVED*** ConfigArrayFactoryInternalSlots
 * @property {Map<string,Plugin>***REMOVED*** additionalPluginPool The map for additional plugins.
 * @property {string***REMOVED*** cwd The path to the current working directory.
 * @property {string | undefined***REMOVED*** resolvePluginsRelativeTo An absolute path the the directory that plugins should be resolved from.
 * @property {Map<string,Rule>***REMOVED*** builtInRules The rules that are built in to ESLint.
 * @property {Object***REMOVED*** [resolver=ModuleResolver] The module resolver object.
 * @property {string***REMOVED*** eslintAllPath The path to the definitions for eslint:all.
 * @property {Function***REMOVED*** getEslintAllConfig Returns the config data for eslint:all.
 * @property {string***REMOVED*** eslintRecommendedPath The path to the definitions for eslint:recommended.
 * @property {Function***REMOVED*** getEslintRecommendedConfig Returns the config data for eslint:recommended.
 */

/**
 * @typedef {Object***REMOVED*** ConfigArrayFactoryLoadingContext
 * @property {string***REMOVED*** filePath The path to the current configuration.
 * @property {string***REMOVED*** matchBasePath The base path to resolve relative paths in `overrides[].files`, `overrides[].excludedFiles`, and `ignorePatterns`.
 * @property {string***REMOVED*** name The name of the current configuration.
 * @property {string***REMOVED*** pluginBasePath The base path to resolve plugins.
 * @property {"config" | "ignore" | "implicit-processor"***REMOVED*** type The type of the current configuration. This is `"config"` in normal. This is `"ignore"` if it came from `.eslintignore`. This is `"implicit-processor"` if it came from legacy file-extension processors.
 */

/**
 * @typedef {Object***REMOVED*** ConfigArrayFactoryLoadingContext
 * @property {string***REMOVED*** filePath The path to the current configuration.
 * @property {string***REMOVED*** matchBasePath The base path to resolve relative paths in `overrides[].files`, `overrides[].excludedFiles`, and `ignorePatterns`.
 * @property {string***REMOVED*** name The name of the current configuration.
 * @property {"config" | "ignore" | "implicit-processor"***REMOVED*** type The type of the current configuration. This is `"config"` in normal. This is `"ignore"` if it came from `.eslintignore`. This is `"implicit-processor"` if it came from legacy file-extension processors.
 */

/** @type {WeakMap<ConfigArrayFactory, ConfigArrayFactoryInternalSlots>***REMOVED*** */
const internalSlotsMap = new WeakMap();

/** @type {WeakMap<object, Plugin>***REMOVED*** */
const normalizedPlugins = new WeakMap();

/**
 * Check if a given string is a file path.
 * @param {string***REMOVED*** nameOrPath A module name or file path.
 * @returns {boolean***REMOVED*** `true` if the `nameOrPath` is a file path.
 */
function isFilePath(nameOrPath) {
    return (
        /^\.{1,2***REMOVED***[/\\]/u.test(nameOrPath) ||
        path.isAbsolute(nameOrPath)
    );
***REMOVED***

/**
 * Convenience wrapper for synchronously reading file contents.
 * @param {string***REMOVED*** filePath The filename to read.
 * @returns {string***REMOVED*** The file contents, with the BOM removed.
 * @private
 */
function readFile(filePath) {
    return fs.readFileSync(filePath, "utf8").replace(/^\ufeff/u, "");
***REMOVED***

/**
 * Loads a YAML configuration from a file.
 * @param {string***REMOVED*** filePath The filename to load.
 * @returns {ConfigData***REMOVED*** The configuration object from the file.
 * @throws {Error***REMOVED*** If the file cannot be read.
 * @private
 */
function loadYAMLConfigFile(filePath) {
    debug(`Loading YAML config file: ${filePath***REMOVED***`);

    // lazy load YAML to improve performance when not used
    const yaml = require("js-yaml");

    try {

        // empty YAML file can be null, so always use
        return yaml.load(readFile(filePath)) || {***REMOVED***;
    ***REMOVED*** catch (e) {
        debug(`Error reading YAML file: ${filePath***REMOVED***`);
        e.message = `Cannot read config file: ${filePath***REMOVED***\nError: ${e.message***REMOVED***`;
        throw e;
    ***REMOVED***
***REMOVED***

/**
 * Loads a JSON configuration from a file.
 * @param {string***REMOVED*** filePath The filename to load.
 * @returns {ConfigData***REMOVED*** The configuration object from the file.
 * @throws {Error***REMOVED*** If the file cannot be read.
 * @private
 */
function loadJSONConfigFile(filePath) {
    debug(`Loading JSON config file: ${filePath***REMOVED***`);

    try {
        return JSON.parse(stripComments(readFile(filePath)));
    ***REMOVED*** catch (e) {
        debug(`Error reading JSON file: ${filePath***REMOVED***`);
        e.message = `Cannot read config file: ${filePath***REMOVED***\nError: ${e.message***REMOVED***`;
        e.messageTemplate = "failed-to-read-json";
        e.messageData = {
            path: filePath,
            message: e.message
        ***REMOVED***;
        throw e;
    ***REMOVED***
***REMOVED***

/**
 * Loads a legacy (.eslintrc) configuration from a file.
 * @param {string***REMOVED*** filePath The filename to load.
 * @returns {ConfigData***REMOVED*** The configuration object from the file.
 * @throws {Error***REMOVED*** If the file cannot be read.
 * @private
 */
function loadLegacyConfigFile(filePath) {
    debug(`Loading legacy config file: ${filePath***REMOVED***`);

    // lazy load YAML to improve performance when not used
    const yaml = require("js-yaml");

    try {
        return yaml.load(stripComments(readFile(filePath))) || /* istanbul ignore next */ {***REMOVED***;
    ***REMOVED*** catch (e) {
        debug("Error reading YAML file: %s\n%o", filePath, e);
        e.message = `Cannot read config file: ${filePath***REMOVED***\nError: ${e.message***REMOVED***`;
        throw e;
    ***REMOVED***
***REMOVED***

/**
 * Loads a JavaScript configuration from a file.
 * @param {string***REMOVED*** filePath The filename to load.
 * @returns {ConfigData***REMOVED*** The configuration object from the file.
 * @throws {Error***REMOVED*** If the file cannot be read.
 * @private
 */
function loadJSConfigFile(filePath) {
    debug(`Loading JS config file: ${filePath***REMOVED***`);
    try {
        return importFresh(filePath);
    ***REMOVED*** catch (e) {
        debug(`Error reading JavaScript file: ${filePath***REMOVED***`);
        e.message = `Cannot read config file: ${filePath***REMOVED***\nError: ${e.message***REMOVED***`;
        throw e;
    ***REMOVED***
***REMOVED***

/**
 * Loads a configuration from a package.json file.
 * @param {string***REMOVED*** filePath The filename to load.
 * @returns {ConfigData***REMOVED*** The configuration object from the file.
 * @throws {Error***REMOVED*** If the file cannot be read.
 * @private
 */
function loadPackageJSONConfigFile(filePath) {
    debug(`Loading package.json config file: ${filePath***REMOVED***`);
    try {
        const packageData = loadJSONConfigFile(filePath);

        if (!Object.hasOwn(packageData, "eslintConfig")) {
            throw Object.assign(
                new Error("package.json file doesn't have 'eslintConfig' field."),
                { code: "ESLINT_CONFIG_FIELD_NOT_FOUND" ***REMOVED***
            );
        ***REMOVED***

        return packageData.eslintConfig;
    ***REMOVED*** catch (e) {
        debug(`Error reading package.json file: ${filePath***REMOVED***`);
        e.message = `Cannot read config file: ${filePath***REMOVED***\nError: ${e.message***REMOVED***`;
        throw e;
    ***REMOVED***
***REMOVED***

/**
 * Loads a `.eslintignore` from a file.
 * @param {string***REMOVED*** filePath The filename to load.
 * @returns {string[]***REMOVED*** The ignore patterns from the file.
 * @throws {Error***REMOVED*** If the file cannot be read.
 * @private
 */
function loadESLintIgnoreFile(filePath) {
    debug(`Loading .eslintignore file: ${filePath***REMOVED***`);

    try {
        return readFile(filePath)
            .split(/\r?\n/gu)
            .filter(line => line.trim() !== "" && !line.startsWith("#"));
    ***REMOVED*** catch (e) {
        debug(`Error reading .eslintignore file: ${filePath***REMOVED***`);
        e.message = `Cannot read .eslintignore file: ${filePath***REMOVED***\nError: ${e.message***REMOVED***`;
        throw e;
    ***REMOVED***
***REMOVED***

/**
 * Creates an error to notify about a missing config to extend from.
 * @param {string***REMOVED*** configName The name of the missing config.
 * @param {string***REMOVED*** importerName The name of the config that imported the missing config
 * @param {string***REMOVED*** messageTemplate The text template to source error strings from.
 * @returns {Error***REMOVED*** The error object to throw
 * @private
 */
function configInvalidError(configName, importerName, messageTemplate) {
    return Object.assign(
        new Error(`Failed to load config "${configName***REMOVED***" to extend from.`),
        {
            messageTemplate,
            messageData: { configName, importerName ***REMOVED***
        ***REMOVED***
    );
***REMOVED***

/**
 * Loads a configuration file regardless of the source. Inspects the file path
 * to determine the correctly way to load the config file.
 * @param {string***REMOVED*** filePath The path to the configuration.
 * @returns {ConfigData|null***REMOVED*** The configuration information.
 * @private
 */
function loadConfigFile(filePath) {
    switch (path.extname(filePath)) {
        case ".js":
        case ".cjs":
            return loadJSConfigFile(filePath);

        case ".json":
            if (path.basename(filePath) === "package.json") {
                return loadPackageJSONConfigFile(filePath);
            ***REMOVED***
            return loadJSONConfigFile(filePath);

        case ".yaml":
        case ".yml":
            return loadYAMLConfigFile(filePath);

        default:
            return loadLegacyConfigFile(filePath);
    ***REMOVED***
***REMOVED***

/**
 * Write debug log.
 * @param {string***REMOVED*** request The requested module name.
 * @param {string***REMOVED*** relativeTo The file path to resolve the request relative to.
 * @param {string***REMOVED*** filePath The resolved file path.
 * @returns {void***REMOVED***
 */
function writeDebugLogForLoading(request, relativeTo, filePath) {
    /* istanbul ignore next */
    if (debug.enabled) {
        let nameAndVersion = null; // eslint-disable-line no-useless-assignment -- known bug in the rule

        try {
            const packageJsonPath = ModuleResolver.resolve(
                `${request***REMOVED***/package.json`,
                relativeTo
            );
            const { version = "unknown" ***REMOVED*** = require(packageJsonPath);

            nameAndVersion = `${request***REMOVED***@${version***REMOVED***`;
        ***REMOVED*** catch (error) {
            debug("package.json was not found:", error.message);
            nameAndVersion = request;
        ***REMOVED***

        debug("Loaded: %s (%s)", nameAndVersion, filePath);
    ***REMOVED***
***REMOVED***

/**
 * Create a new context with default values.
 * @param {ConfigArrayFactoryInternalSlots***REMOVED*** slots The internal slots.
 * @param {"config" | "ignore" | "implicit-processor" | undefined***REMOVED*** providedType The type of the current configuration. Default is `"config"`.
 * @param {string | undefined***REMOVED*** providedName The name of the current configuration. Default is the relative path from `cwd` to `filePath`.
 * @param {string | undefined***REMOVED*** providedFilePath The path to the current configuration. Default is empty string.
 * @param {string | undefined***REMOVED*** providedMatchBasePath The type of the current configuration. Default is the directory of `filePath` or `cwd`.
 * @returns {ConfigArrayFactoryLoadingContext***REMOVED*** The created context.
 */
function createContext(
    { cwd, resolvePluginsRelativeTo ***REMOVED***,
    providedType,
    providedName,
    providedFilePath,
    providedMatchBasePath
) {
    const filePath = providedFilePath
        ? path.resolve(cwd, providedFilePath)
        : "";
    const matchBasePath =
        (providedMatchBasePath && path.resolve(cwd, providedMatchBasePath)) ||
        (filePath && path.dirname(filePath)) ||
        cwd;
    const name =
        providedName ||
        (filePath && path.relative(cwd, filePath)) ||
        "";
    const pluginBasePath =
        resolvePluginsRelativeTo ||
        (filePath && path.dirname(filePath)) ||
        cwd;
    const type = providedType || "config";

    return { filePath, matchBasePath, name, pluginBasePath, type ***REMOVED***;
***REMOVED***

/**
 * Normalize a given plugin.
 * - Ensure the object to have four properties: configs, environments, processors, and rules.
 * - Ensure the object to not have other properties.
 * @param {Plugin***REMOVED*** plugin The plugin to normalize.
 * @returns {Plugin***REMOVED*** The normalized plugin.
 */
function normalizePlugin(plugin) {

    // first check the cache
    let normalizedPlugin = normalizedPlugins.get(plugin);

    if (normalizedPlugin) {
        return normalizedPlugin;
    ***REMOVED***

    normalizedPlugin = {
        configs: plugin.configs || {***REMOVED***,
        environments: plugin.environments || {***REMOVED***,
        processors: plugin.processors || {***REMOVED***,
        rules: plugin.rules || {***REMOVED***
    ***REMOVED***;

    // save the reference for later
    normalizedPlugins.set(plugin, normalizedPlugin);

    return normalizedPlugin;
***REMOVED***

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

/**
 * The factory of `ConfigArray` objects.
 */
class ConfigArrayFactory {

    /**
     * Initialize this instance.
     * @param {ConfigArrayFactoryOptions***REMOVED*** [options] The map for additional plugins.
     */
    constructor({
        additionalPluginPool = new Map(),
        cwd = process.cwd(),
        resolvePluginsRelativeTo,
        builtInRules,
        resolver = ModuleResolver,
        eslintAllPath,
        getEslintAllConfig,
        eslintRecommendedPath,
        getEslintRecommendedConfig
    ***REMOVED*** = {***REMOVED***) {
        internalSlotsMap.set(this, {
            additionalPluginPool,
            cwd,
            resolvePluginsRelativeTo:
                resolvePluginsRelativeTo &&
                path.resolve(cwd, resolvePluginsRelativeTo),
            builtInRules,
            resolver,
            eslintAllPath,
            getEslintAllConfig,
            eslintRecommendedPath,
            getEslintRecommendedConfig
        ***REMOVED***);
    ***REMOVED***

    /**
     * Create `ConfigArray` instance from a config data.
     * @param {ConfigData|null***REMOVED*** configData The config data to create.
     * @param {Object***REMOVED*** [options] The options.
     * @param {string***REMOVED*** [options.basePath] The base path to resolve relative paths in `overrides[].files`, `overrides[].excludedFiles`, and `ignorePatterns`.
     * @param {string***REMOVED*** [options.filePath] The path to this config data.
     * @param {string***REMOVED*** [options.name] The config name.
     * @returns {ConfigArray***REMOVED*** Loaded config.
     */
    create(configData, { basePath, filePath, name ***REMOVED*** = {***REMOVED***) {
        if (!configData) {
            return new ConfigArray();
        ***REMOVED***

        const slots = internalSlotsMap.get(this);
        const ctx = createContext(slots, "config", name, filePath, basePath);
        const elements = this._normalizeConfigData(configData, ctx);

        return new ConfigArray(...elements);
    ***REMOVED***

    /**
     * Load a config file.
     * @param {string***REMOVED*** filePath The path to a config file.
     * @param {Object***REMOVED*** [options] The options.
     * @param {string***REMOVED*** [options.basePath] The base path to resolve relative paths in `overrides[].files`, `overrides[].excludedFiles`, and `ignorePatterns`.
     * @param {string***REMOVED*** [options.name] The config name.
     * @returns {ConfigArray***REMOVED*** Loaded config.
     */
    loadFile(filePath, { basePath, name ***REMOVED*** = {***REMOVED***) {
        const slots = internalSlotsMap.get(this);
        const ctx = createContext(slots, "config", name, filePath, basePath);

        return new ConfigArray(...this._loadConfigData(ctx));
    ***REMOVED***

    /**
     * Load the config file on a given directory if exists.
     * @param {string***REMOVED*** directoryPath The path to a directory.
     * @param {Object***REMOVED*** [options] The options.
     * @param {string***REMOVED*** [options.basePath] The base path to resolve relative paths in `overrides[].files`, `overrides[].excludedFiles`, and `ignorePatterns`.
     * @param {string***REMOVED*** [options.name] The config name.
     * @throws {Error***REMOVED*** If the config file is invalid.
     * @returns {ConfigArray***REMOVED*** Loaded config. An empty `ConfigArray` if any config doesn't exist.
     */
    loadInDirectory(directoryPath, { basePath, name ***REMOVED*** = {***REMOVED***) {
        const slots = internalSlotsMap.get(this);

        for (const filename of configFilenames) {
            const ctx = createContext(
                slots,
                "config",
                name,
                path.join(directoryPath, filename),
                basePath
            );

            if (fs.existsSync(ctx.filePath) && fs.statSync(ctx.filePath).isFile()) {
                let configData;

                try {
                    configData = loadConfigFile(ctx.filePath);
                ***REMOVED*** catch (error) {
                    if (!error || error.code !== "ESLINT_CONFIG_FIELD_NOT_FOUND") {
                        throw error;
                    ***REMOVED***
                ***REMOVED***

                if (configData) {
                    debug(`Config file found: ${ctx.filePath***REMOVED***`);
                    return new ConfigArray(
                        ...this._normalizeConfigData(configData, ctx)
                    );
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        debug(`Config file not found on ${directoryPath***REMOVED***`);
        return new ConfigArray();
    ***REMOVED***

    /**
     * Check if a config file on a given directory exists or not.
     * @param {string***REMOVED*** directoryPath The path to a directory.
     * @returns {string | null***REMOVED*** The path to the found config file. If not found then null.
     */
    static getPathToConfigFileInDirectory(directoryPath) {
        for (const filename of configFilenames) {
            const filePath = path.join(directoryPath, filename);

            if (fs.existsSync(filePath)) {
                if (filename === "package.json") {
                    try {
                        loadPackageJSONConfigFile(filePath);
                        return filePath;
                    ***REMOVED*** catch { /* ignore */ ***REMOVED***
                ***REMOVED*** else {
                    return filePath;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        return null;
    ***REMOVED***

    /**
     * Load `.eslintignore` file.
     * @param {string***REMOVED*** filePath The path to a `.eslintignore` file to load.
     * @returns {ConfigArray***REMOVED*** Loaded config. An empty `ConfigArray` if any config doesn't exist.
     */
    loadESLintIgnore(filePath) {
        const slots = internalSlotsMap.get(this);
        const ctx = createContext(
            slots,
            "ignore",
            void 0,
            filePath,
            slots.cwd
        );
        const ignorePatterns = loadESLintIgnoreFile(ctx.filePath);

        return new ConfigArray(
            ...this._normalizeESLintIgnoreData(ignorePatterns, ctx)
        );
    ***REMOVED***

    /**
     * Load `.eslintignore` file in the current working directory.
     * @returns {ConfigArray***REMOVED*** Loaded config. An empty `ConfigArray` if any config doesn't exist.
     * @throws {Error***REMOVED*** If the ignore file is invalid.
     */
    loadDefaultESLintIgnore() {
        const slots = internalSlotsMap.get(this);
        const eslintIgnorePath = path.resolve(slots.cwd, ".eslintignore");
        const packageJsonPath = path.resolve(slots.cwd, "package.json");

        if (fs.existsSync(eslintIgnorePath)) {
            return this.loadESLintIgnore(eslintIgnorePath);
        ***REMOVED***
        if (fs.existsSync(packageJsonPath)) {
            const data = loadJSONConfigFile(packageJsonPath);

            if (Object.hasOwn(data, "eslintIgnore")) {
                if (!Array.isArray(data.eslintIgnore)) {
                    throw new Error("Package.json eslintIgnore property requires an array of paths");
                ***REMOVED***
                const ctx = createContext(
                    slots,
                    "ignore",
                    "eslintIgnore in package.json",
                    packageJsonPath,
                    slots.cwd
                );

                return new ConfigArray(
                    ...this._normalizeESLintIgnoreData(data.eslintIgnore, ctx)
                );
            ***REMOVED***
        ***REMOVED***

        return new ConfigArray();
    ***REMOVED***

    /**
     * Load a given config file.
     * @param {ConfigArrayFactoryLoadingContext***REMOVED*** ctx The loading context.
     * @returns {IterableIterator<ConfigArrayElement>***REMOVED*** Loaded config.
     * @private
     */
    _loadConfigData(ctx) {
        return this._normalizeConfigData(loadConfigFile(ctx.filePath), ctx);
    ***REMOVED***

    /**
     * Normalize a given `.eslintignore` data to config array elements.
     * @param {string[]***REMOVED*** ignorePatterns The patterns to ignore files.
     * @param {ConfigArrayFactoryLoadingContext***REMOVED*** ctx The loading context.
     * @returns {IterableIterator<ConfigArrayElement>***REMOVED*** The normalized config.
     * @private
     */
    *_normalizeESLintIgnoreData(ignorePatterns, ctx) {
        const elements = this._normalizeObjectConfigData(
            { ignorePatterns ***REMOVED***,
            ctx
        );

        // Set `ignorePattern.loose` flag for backward compatibility.
        for (const element of elements) {
            if (element.ignorePattern) {
                element.ignorePattern.loose = true;
            ***REMOVED***
            yield element;
        ***REMOVED***
    ***REMOVED***

    /**
     * Normalize a given config to an array.
     * @param {ConfigData***REMOVED*** configData The config data to normalize.
     * @param {ConfigArrayFactoryLoadingContext***REMOVED*** ctx The loading context.
     * @returns {IterableIterator<ConfigArrayElement>***REMOVED*** The normalized config.
     * @private
     */
    _normalizeConfigData(configData, ctx) {
        const validator = new ConfigValidator();

        validator.validateConfigSchema(configData, ctx.name || ctx.filePath);
        return this._normalizeObjectConfigData(configData, ctx);
    ***REMOVED***

    /**
     * Normalize a given config to an array.
     * @param {ConfigData|OverrideConfigData***REMOVED*** configData The config data to normalize.
     * @param {ConfigArrayFactoryLoadingContext***REMOVED*** ctx The loading context.
     * @returns {IterableIterator<ConfigArrayElement>***REMOVED*** The normalized config.
     * @private
     */
    *_normalizeObjectConfigData(configData, ctx) {
        const { files, excludedFiles, ...configBody ***REMOVED*** = configData;
        const criteria = OverrideTester.create(
            files,
            excludedFiles,
            ctx.matchBasePath
        );
        const elements = this._normalizeObjectConfigDataBody(configBody, ctx);

        // Apply the criteria to every element.
        for (const element of elements) {

            /*
             * Merge the criteria.
             * This is for the `overrides` entries that came from the
             * configurations of `overrides[].extends`.
             */
            element.criteria = OverrideTester.and(criteria, element.criteria);

            /*
             * Remove `root` property to ignore `root` settings which came from
             * `extends` in `overrides`.
             */
            if (element.criteria) {
                element.root = void 0;
            ***REMOVED***

            yield element;
        ***REMOVED***
    ***REMOVED***

    /**
     * Normalize a given config to an array.
     * @param {ConfigData***REMOVED*** configData The config data to normalize.
     * @param {ConfigArrayFactoryLoadingContext***REMOVED*** ctx The loading context.
     * @returns {IterableIterator<ConfigArrayElement>***REMOVED*** The normalized config.
     * @private
     */
    *_normalizeObjectConfigDataBody(
        {
            env,
            extends: extend,
            globals,
            ignorePatterns,
            noInlineConfig,
            parser: parserName,
            parserOptions,
            plugins: pluginList,
            processor,
            reportUnusedDisableDirectives,
            root,
            rules,
            settings,
            overrides: overrideList = []
        ***REMOVED***,
        ctx
    ) {
        const extendList = Array.isArray(extend) ? extend : [extend];
        const ignorePattern = ignorePatterns && new IgnorePattern(
            Array.isArray(ignorePatterns) ? ignorePatterns : [ignorePatterns],
            ctx.matchBasePath
        );

        // Flatten `extends`.
        for (const extendName of extendList.filter(Boolean)) {
            yield* this._loadExtends(extendName, ctx);
        ***REMOVED***

        // Load parser & plugins.
        const parser = parserName && this._loadParser(parserName, ctx);
        const plugins = pluginList && this._loadPlugins(pluginList, ctx);

        // Yield pseudo config data for file extension processors.
        if (plugins) {
            yield* this._takeFileExtensionProcessors(plugins, ctx);
        ***REMOVED***

        // Yield the config data except `extends` and `overrides`.
        yield {

            // Debug information.
            type: ctx.type,
            name: ctx.name,
            filePath: ctx.filePath,

            // Config data.
            criteria: null,
            env,
            globals,
            ignorePattern,
            noInlineConfig,
            parser,
            parserOptions,
            plugins,
            processor,
            reportUnusedDisableDirectives,
            root,
            rules,
            settings
        ***REMOVED***;

        // Flatten `overries`.
        for (let i = 0; i < overrideList.length; ++i) {
            yield* this._normalizeObjectConfigData(
                overrideList[i],
                { ...ctx, name: `${ctx.name***REMOVED***#overrides[${i***REMOVED***]` ***REMOVED***
            );
        ***REMOVED***
    ***REMOVED***

    /**
     * Load configs of an element in `extends`.
     * @param {string***REMOVED*** extendName The name of a base config.
     * @param {ConfigArrayFactoryLoadingContext***REMOVED*** ctx The loading context.
     * @returns {IterableIterator<ConfigArrayElement>***REMOVED*** The normalized config.
     * @throws {Error***REMOVED*** If the extended config file can't be loaded.
     * @private
     */
    _loadExtends(extendName, ctx) {
        debug("Loading {extends:%j***REMOVED*** relative to %s", extendName, ctx.filePath);
        try {
            if (extendName.startsWith("eslint:")) {
                return this._loadExtendedBuiltInConfig(extendName, ctx);
            ***REMOVED***
            if (extendName.startsWith("plugin:")) {
                return this._loadExtendedPluginConfig(extendName, ctx);
            ***REMOVED***
            return this._loadExtendedShareableConfig(extendName, ctx);
        ***REMOVED*** catch (error) {
            error.message += `\nReferenced from: ${ctx.filePath || ctx.name***REMOVED***`;
            throw error;
        ***REMOVED***
    ***REMOVED***

    /**
     * Load configs of an element in `extends`.
     * @param {string***REMOVED*** extendName The name of a base config.
     * @param {ConfigArrayFactoryLoadingContext***REMOVED*** ctx The loading context.
     * @returns {IterableIterator<ConfigArrayElement>***REMOVED*** The normalized config.
     * @throws {Error***REMOVED*** If the extended config file can't be loaded.
     * @private
     */
    _loadExtendedBuiltInConfig(extendName, ctx) {
        const {
            eslintAllPath,
            getEslintAllConfig,
            eslintRecommendedPath,
            getEslintRecommendedConfig
        ***REMOVED*** = internalSlotsMap.get(this);

        if (extendName === "eslint:recommended") {
            const name = `${ctx.name***REMOVED*** » ${extendName***REMOVED***`;

            if (getEslintRecommendedConfig) {
                if (typeof getEslintRecommendedConfig !== "function") {
                    throw new Error(`getEslintRecommendedConfig must be a function instead of '${getEslintRecommendedConfig***REMOVED***'`);
                ***REMOVED***
                return this._normalizeConfigData(getEslintRecommendedConfig(), { ...ctx, name, filePath: "" ***REMOVED***);
            ***REMOVED***
            return this._loadConfigData({
                ...ctx,
                name,
                filePath: eslintRecommendedPath
            ***REMOVED***);
        ***REMOVED***
        if (extendName === "eslint:all") {
            const name = `${ctx.name***REMOVED*** » ${extendName***REMOVED***`;

            if (getEslintAllConfig) {
                if (typeof getEslintAllConfig !== "function") {
                    throw new Error(`getEslintAllConfig must be a function instead of '${getEslintAllConfig***REMOVED***'`);
                ***REMOVED***
                return this._normalizeConfigData(getEslintAllConfig(), { ...ctx, name, filePath: "" ***REMOVED***);
            ***REMOVED***
            return this._loadConfigData({
                ...ctx,
                name,
                filePath: eslintAllPath
            ***REMOVED***);
        ***REMOVED***

        throw configInvalidError(extendName, ctx.name, "extend-config-missing");
    ***REMOVED***

    /**
     * Load configs of an element in `extends`.
     * @param {string***REMOVED*** extendName The name of a base config.
     * @param {ConfigArrayFactoryLoadingContext***REMOVED*** ctx The loading context.
     * @returns {IterableIterator<ConfigArrayElement>***REMOVED*** The normalized config.
     * @throws {Error***REMOVED*** If the extended config file can't be loaded.
     * @private
     */
    _loadExtendedPluginConfig(extendName, ctx) {
        const slashIndex = extendName.lastIndexOf("/");

        if (slashIndex === -1) {
            throw configInvalidError(extendName, ctx.filePath, "plugin-invalid");
        ***REMOVED***

        const pluginName = extendName.slice("plugin:".length, slashIndex);
        const configName = extendName.slice(slashIndex + 1);

        if (isFilePath(pluginName)) {
            throw new Error("'extends' cannot use a file path for plugins.");
        ***REMOVED***

        const plugin = this._loadPlugin(pluginName, ctx);
        const configData =
            plugin.definition &&
            plugin.definition.configs[configName];

        if (configData) {
            return this._normalizeConfigData(configData, {
                ...ctx,
                filePath: plugin.filePath || ctx.filePath,
                name: `${ctx.name***REMOVED*** » plugin:${plugin.id***REMOVED***/${configName***REMOVED***`
            ***REMOVED***);
        ***REMOVED***

        throw plugin.error || configInvalidError(extendName, ctx.filePath, "extend-config-missing");
    ***REMOVED***

    /**
     * Load configs of an element in `extends`.
     * @param {string***REMOVED*** extendName The name of a base config.
     * @param {ConfigArrayFactoryLoadingContext***REMOVED*** ctx The loading context.
     * @returns {IterableIterator<ConfigArrayElement>***REMOVED*** The normalized config.
     * @throws {Error***REMOVED*** If the extended config file can't be loaded.
     * @private
     */
    _loadExtendedShareableConfig(extendName, ctx) {
        const { cwd, resolver ***REMOVED*** = internalSlotsMap.get(this);
        const relativeTo = ctx.filePath || path.join(cwd, "__placeholder__.js");
        let request;

        if (isFilePath(extendName)) {
            request = extendName;
        ***REMOVED*** else if (extendName.startsWith(".")) {
            request = `./${extendName***REMOVED***`; // For backward compatibility. A ton of tests depended on this behavior.
        ***REMOVED*** else {
            request = naming.normalizePackageName(
                extendName,
                "eslint-config"
            );
        ***REMOVED***

        let filePath;

        try {
            filePath = resolver.resolve(request, relativeTo);
        ***REMOVED*** catch (error) {
            /* istanbul ignore else */
            if (error && error.code === "MODULE_NOT_FOUND") {
                throw configInvalidError(extendName, ctx.filePath, "extend-config-missing");
            ***REMOVED***
            throw error;
        ***REMOVED***

        writeDebugLogForLoading(request, relativeTo, filePath);
        return this._loadConfigData({
            ...ctx,
            filePath,
            name: `${ctx.name***REMOVED*** » ${request***REMOVED***`
        ***REMOVED***);
    ***REMOVED***

    /**
     * Load given plugins.
     * @param {string[]***REMOVED*** names The plugin names to load.
     * @param {ConfigArrayFactoryLoadingContext***REMOVED*** ctx The loading context.
     * @returns {Record<string,DependentPlugin>***REMOVED*** The loaded parser.
     * @private
     */
    _loadPlugins(names, ctx) {
        return names.reduce((map, name) => {
            if (isFilePath(name)) {
                throw new Error("Plugins array cannot includes file paths.");
            ***REMOVED***
            const plugin = this._loadPlugin(name, ctx);

            map[plugin.id] = plugin;

            return map;
        ***REMOVED***, {***REMOVED***);
    ***REMOVED***

    /**
     * Load a given parser.
     * @param {string***REMOVED*** nameOrPath The package name or the path to a parser file.
     * @param {ConfigArrayFactoryLoadingContext***REMOVED*** ctx The loading context.
     * @returns {DependentParser***REMOVED*** The loaded parser.
     */
    _loadParser(nameOrPath, ctx) {
        debug("Loading parser %j from %s", nameOrPath, ctx.filePath);

        const { cwd, resolver ***REMOVED*** = internalSlotsMap.get(this);
        const relativeTo = ctx.filePath || path.join(cwd, "__placeholder__.js");

        try {
            const filePath = resolver.resolve(nameOrPath, relativeTo);

            writeDebugLogForLoading(nameOrPath, relativeTo, filePath);

            return new ConfigDependency({
                definition: require(filePath),
                filePath,
                id: nameOrPath,
                importerName: ctx.name,
                importerPath: ctx.filePath
            ***REMOVED***);
        ***REMOVED*** catch (error) {

            // If the parser name is "espree", load the espree of ESLint.
            if (nameOrPath === "espree") {
                debug("Fallback espree.");
                return new ConfigDependency({
                    definition: require("espree"),
                    filePath: require.resolve("espree"),
                    id: nameOrPath,
                    importerName: ctx.name,
                    importerPath: ctx.filePath
                ***REMOVED***);
            ***REMOVED***

            debug("Failed to load parser '%s' declared in '%s'.", nameOrPath, ctx.name);
            error.message = `Failed to load parser '${nameOrPath***REMOVED***' declared in '${ctx.name***REMOVED***': ${error.message***REMOVED***`;

            return new ConfigDependency({
                error,
                id: nameOrPath,
                importerName: ctx.name,
                importerPath: ctx.filePath
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***

    /**
     * Load a given plugin.
     * @param {string***REMOVED*** name The plugin name to load.
     * @param {ConfigArrayFactoryLoadingContext***REMOVED*** ctx The loading context.
     * @returns {DependentPlugin***REMOVED*** The loaded plugin.
     * @private
     */
    _loadPlugin(name, ctx) {
        debug("Loading plugin %j from %s", name, ctx.filePath);

        const { additionalPluginPool, resolver ***REMOVED*** = internalSlotsMap.get(this);
        const request = naming.normalizePackageName(name, "eslint-plugin");
        const id = naming.getShorthandName(request, "eslint-plugin");
        const relativeTo = path.join(ctx.pluginBasePath, "__placeholder__.js");

        if (name.match(/\s+/u)) {
            const error = Object.assign(
                new Error(`Whitespace found in plugin name '${name***REMOVED***'`),
                {
                    messageTemplate: "whitespace-found",
                    messageData: { pluginName: request ***REMOVED***
                ***REMOVED***
            );

            return new ConfigDependency({
                error,
                id,
                importerName: ctx.name,
                importerPath: ctx.filePath
            ***REMOVED***);
        ***REMOVED***

        // Check for additional pool.
        const plugin =
            additionalPluginPool.get(request) ||
            additionalPluginPool.get(id);

        if (plugin) {
            return new ConfigDependency({
                definition: normalizePlugin(plugin),
                original: plugin,
                filePath: "", // It's unknown where the plugin came from.
                id,
                importerName: ctx.name,
                importerPath: ctx.filePath
            ***REMOVED***);
        ***REMOVED***

        let filePath;
        let error;

        try {
            filePath = resolver.resolve(request, relativeTo);
        ***REMOVED*** catch (resolveError) {
            error = resolveError;
            /* istanbul ignore else */
            if (error && error.code === "MODULE_NOT_FOUND") {
                error.messageTemplate = "plugin-missing";
                error.messageData = {
                    pluginName: request,
                    resolvePluginsRelativeTo: ctx.pluginBasePath,
                    importerName: ctx.name
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***

        if (filePath) {
            try {
                writeDebugLogForLoading(request, relativeTo, filePath);

                const startTime = Date.now();
                const pluginDefinition = require(filePath);

                debug(`Plugin ${filePath***REMOVED*** loaded in: ${Date.now() - startTime***REMOVED***ms`);

                return new ConfigDependency({
                    definition: normalizePlugin(pluginDefinition),
                    original: pluginDefinition,
                    filePath,
                    id,
                    importerName: ctx.name,
                    importerPath: ctx.filePath
                ***REMOVED***);
            ***REMOVED*** catch (loadError) {
                error = loadError;
            ***REMOVED***
        ***REMOVED***

        debug("Failed to load plugin '%s' declared in '%s'.", name, ctx.name);
        error.message = `Failed to load plugin '${name***REMOVED***' declared in '${ctx.name***REMOVED***': ${error.message***REMOVED***`;
        return new ConfigDependency({
            error,
            id,
            importerName: ctx.name,
            importerPath: ctx.filePath
        ***REMOVED***);
    ***REMOVED***

    /**
     * Take file expression processors as config array elements.
     * @param {Record<string,DependentPlugin>***REMOVED*** plugins The plugin definitions.
     * @param {ConfigArrayFactoryLoadingContext***REMOVED*** ctx The loading context.
     * @returns {IterableIterator<ConfigArrayElement>***REMOVED*** The config array elements of file expression processors.
     * @private
     */
    *_takeFileExtensionProcessors(plugins, ctx) {
        for (const pluginId of Object.keys(plugins)) {
            const processors =
                plugins[pluginId] &&
                plugins[pluginId].definition &&
                plugins[pluginId].definition.processors;

            if (!processors) {
                continue;
            ***REMOVED***

            for (const processorId of Object.keys(processors)) {
                if (processorId.startsWith(".")) {
                    yield* this._normalizeObjectConfigData(
                        {
                            files: [`*${processorId***REMOVED***`],
                            processor: `${pluginId***REMOVED***/${processorId***REMOVED***`
                        ***REMOVED***,
                        {
                            ...ctx,
                            type: "implicit-processor",
                            name: `${ctx.name***REMOVED***#processors["${pluginId***REMOVED***/${processorId***REMOVED***"]`
                        ***REMOVED***
                    );
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
***REMOVED***

export {
    ConfigArrayFactory,
    createContext,
    loadConfigFile
***REMOVED***;
