// @eslint/eslintrc/lib/flat-compat.js
/**
 * @fileoverview Compatibility class for flat config.
 * @author Nicholas C. Zakas
 */

//-----------------------------------------------------------------------------
// Requirements
//-----------------------------------------------------------------------------

import createDebug from "debug";
import path from "node:path";

import environments from "../conf/environments.js";
import { ConfigArrayFactory ***REMOVED*** from "./config-array-factory.js";

//-----------------------------------------------------------------------------
// Helpers
//-----------------------------------------------------------------------------

/** @typedef {import("../../shared/types").Environment***REMOVED*** Environment */
/** @typedef {import("../../shared/types").Processor***REMOVED*** Processor */

const debug = createDebug("eslintrc:flat-compat");
const cafactory = Symbol("cafactory");

/**
 * Translates an ESLintRC-style config object into a flag-config-style config
 * object.
 * @param {Object***REMOVED*** eslintrcConfig An ESLintRC-style config object.
 * @param {Object***REMOVED*** options Options to help translate the config.
 * @param {string***REMOVED*** options.resolveConfigRelativeTo To the directory to resolve
 *      configs from.
 * @param {string***REMOVED*** options.resolvePluginsRelativeTo The directory to resolve
 *      plugins from.
 * @param {ReadOnlyMap<string,Environment>***REMOVED*** options.pluginEnvironments A map of plugin environment
 *      names to objects.
 * @param {ReadOnlyMap<string,Processor>***REMOVED*** options.pluginProcessors A map of plugin processor
 *      names to objects.
 * @returns {Object***REMOVED*** A flag-config-style config object.
 * @throws {Error***REMOVED*** If a plugin or environment cannot be resolved.
 */
function translateESLintRC(eslintrcConfig, {
    resolveConfigRelativeTo,
    resolvePluginsRelativeTo,
    pluginEnvironments,
    pluginProcessors
***REMOVED***) {

    const flatConfig = {***REMOVED***;
    const configs = [];
    const languageOptions = {***REMOVED***;
    const linterOptions = {***REMOVED***;
    const keysToCopy = ["settings", "rules", "processor"];
    const languageOptionsKeysToCopy = ["globals", "parser", "parserOptions"];
    const linterOptionsKeysToCopy = ["noInlineConfig", "reportUnusedDisableDirectives"];

    // copy over simple translations
    for (const key of keysToCopy) {
        if (key in eslintrcConfig && typeof eslintrcConfig[key] !== "undefined") {
            flatConfig[key] = eslintrcConfig[key];
        ***REMOVED***
    ***REMOVED***

    // copy over languageOptions
    for (const key of languageOptionsKeysToCopy) {
        if (key in eslintrcConfig && typeof eslintrcConfig[key] !== "undefined") {

            // create the languageOptions key in the flat config
            flatConfig.languageOptions = languageOptions;

            if (key === "parser") {
                debug(`Resolving parser '${languageOptions[key]***REMOVED***' relative to ${resolveConfigRelativeTo***REMOVED***`);

                if (eslintrcConfig[key].error) {
                    throw eslintrcConfig[key].error;
                ***REMOVED***

                languageOptions[key] = eslintrcConfig[key].definition;
                continue;
            ***REMOVED***

            // clone any object values that are in the eslintrc config
            if (eslintrcConfig[key] && typeof eslintrcConfig[key] === "object") {
                languageOptions[key] = {
                    ...eslintrcConfig[key]
                ***REMOVED***;
            ***REMOVED*** else {
                languageOptions[key] = eslintrcConfig[key];
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    // copy over linterOptions
    for (const key of linterOptionsKeysToCopy) {
        if (key in eslintrcConfig && typeof eslintrcConfig[key] !== "undefined") {
            flatConfig.linterOptions = linterOptions;
            linterOptions[key] = eslintrcConfig[key];
        ***REMOVED***
    ***REMOVED***

    // move ecmaVersion a level up
    if (languageOptions.parserOptions) {

        if ("ecmaVersion" in languageOptions.parserOptions) {
            languageOptions.ecmaVersion = languageOptions.parserOptions.ecmaVersion;
            delete languageOptions.parserOptions.ecmaVersion;
        ***REMOVED***

        if ("sourceType" in languageOptions.parserOptions) {
            languageOptions.sourceType = languageOptions.parserOptions.sourceType;
            delete languageOptions.parserOptions.sourceType;
        ***REMOVED***

        // check to see if we even need parserOptions anymore and remove it if not
        if (Object.keys(languageOptions.parserOptions).length === 0) {
            delete languageOptions.parserOptions;
        ***REMOVED***
    ***REMOVED***

    // overrides
    if (eslintrcConfig.criteria) {
        flatConfig.files = [absoluteFilePath => eslintrcConfig.criteria.test(absoluteFilePath)];
    ***REMOVED***

    // translate plugins
    if (eslintrcConfig.plugins && typeof eslintrcConfig.plugins === "object") {
        debug(`Translating plugins: ${eslintrcConfig.plugins***REMOVED***`);

        flatConfig.plugins = {***REMOVED***;

        for (const pluginName of Object.keys(eslintrcConfig.plugins)) {

            debug(`Translating plugin: ${pluginName***REMOVED***`);
            debug(`Resolving plugin '${pluginName***REMOVED*** relative to ${resolvePluginsRelativeTo***REMOVED***`);

            const { original: plugin, error ***REMOVED*** = eslintrcConfig.plugins[pluginName];

            if (error) {
                throw error;
            ***REMOVED***

            flatConfig.plugins[pluginName] = plugin;

            // create a config for any processors
            if (plugin.processors) {
                for (const processorName of Object.keys(plugin.processors)) {
                    if (processorName.startsWith(".")) {
                        debug(`Assigning processor: ${pluginName***REMOVED***/${processorName***REMOVED***`);

                        configs.unshift({
                            files: [`**/*${processorName***REMOVED***`],
                            processor: pluginProcessors.get(`${pluginName***REMOVED***/${processorName***REMOVED***`)
                        ***REMOVED***);
                    ***REMOVED***

                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    // translate env - must come after plugins
    if (eslintrcConfig.env && typeof eslintrcConfig.env === "object") {
        for (const envName of Object.keys(eslintrcConfig.env)) {

            // only add environments that are true
            if (eslintrcConfig.env[envName]) {
                debug(`Translating environment: ${envName***REMOVED***`);

                if (environments.has(envName)) {

                    // built-in environments should be defined first
                    configs.unshift(...translateESLintRC({
                        criteria: eslintrcConfig.criteria,
                        ...environments.get(envName)
                    ***REMOVED***, {
                        resolveConfigRelativeTo,
                        resolvePluginsRelativeTo
                    ***REMOVED***));
                ***REMOVED*** else if (pluginEnvironments.has(envName)) {

                    // if the environment comes from a plugin, it should come after the plugin config
                    configs.push(...translateESLintRC({
                        criteria: eslintrcConfig.criteria,
                        ...pluginEnvironments.get(envName)
                    ***REMOVED***, {
                        resolveConfigRelativeTo,
                        resolvePluginsRelativeTo
                    ***REMOVED***));
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    // only add if there are actually keys in the config
    if (Object.keys(flatConfig).length > 0) {
        configs.push(flatConfig);
    ***REMOVED***

    return configs;
***REMOVED***


//-----------------------------------------------------------------------------
// Exports
//-----------------------------------------------------------------------------

/**
 * A compatibility class for working with configs.
 */
class FlatCompat {

    constructor({
        baseDirectory = process.cwd(),
        resolvePluginsRelativeTo = baseDirectory,
        recommendedConfig,
        allConfig
    ***REMOVED*** = {***REMOVED***) {
        this.baseDirectory = baseDirectory;
        this.resolvePluginsRelativeTo = resolvePluginsRelativeTo;
        this[cafactory] = new ConfigArrayFactory({
            cwd: baseDirectory,
            resolvePluginsRelativeTo,
            getEslintAllConfig() {

                if (!allConfig) {
                    throw new TypeError("Missing parameter 'allConfig' in FlatCompat constructor.");
                ***REMOVED***

                return allConfig;
            ***REMOVED***,
            getEslintRecommendedConfig() {

                if (!recommendedConfig) {
                    throw new TypeError("Missing parameter 'recommendedConfig' in FlatCompat constructor.");
                ***REMOVED***

                return recommendedConfig;
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***

    /**
     * Translates an ESLintRC-style config into a flag-config-style config.
     * @param {Object***REMOVED*** eslintrcConfig The ESLintRC-style config object.
     * @returns {Object***REMOVED*** A flag-config-style config object.
     */
    config(eslintrcConfig) {
        const eslintrcArray = this[cafactory].create(eslintrcConfig, {
            basePath: this.baseDirectory
        ***REMOVED***);

        const flatArray = [];
        let hasIgnorePatterns = false;

        eslintrcArray.forEach(configData => {
            if (configData.type === "config") {
                hasIgnorePatterns = hasIgnorePatterns || configData.ignorePattern;
                flatArray.push(...translateESLintRC(configData, {
                    resolveConfigRelativeTo: path.join(this.baseDirectory, "__placeholder.js"),
                    resolvePluginsRelativeTo: path.join(this.resolvePluginsRelativeTo, "__placeholder.js"),
                    pluginEnvironments: eslintrcArray.pluginEnvironments,
                    pluginProcessors: eslintrcArray.pluginProcessors
                ***REMOVED***));
            ***REMOVED***
        ***REMOVED***);

        // combine ignorePatterns to emulate ESLintRC behavior better
        if (hasIgnorePatterns) {
            flatArray.unshift({
                ignores: [filePath => {

                    // Compute the final config for this file.
                    // This filters config array elements by `files`/`excludedFiles` then merges the elements.
                    const finalConfig = eslintrcArray.extractConfig(filePath);

                    // Test the `ignorePattern` properties of the final config.
                    return Boolean(finalConfig.ignores) && finalConfig.ignores(filePath);
                ***REMOVED***]
            ***REMOVED***);
        ***REMOVED***

        return flatArray;
    ***REMOVED***

    /**
     * Translates the `env` section of an ESLintRC-style config.
     * @param {Object***REMOVED*** envConfig The `env` section of an ESLintRC config.
     * @returns {Object[]***REMOVED*** An array of flag-config objects representing the environments.
     */
    env(envConfig) {
        return this.config({
            env: envConfig
        ***REMOVED***);
    ***REMOVED***

    /**
     * Translates the `extends` section of an ESLintRC-style config.
     * @param {...string***REMOVED*** configsToExtend The names of the configs to load.
     * @returns {Object[]***REMOVED*** An array of flag-config objects representing the config.
     */
    extends(...configsToExtend) {
        return this.config({
            extends: configsToExtend
        ***REMOVED***);
    ***REMOVED***

    /**
     * Translates the `plugins` section of an ESLintRC-style config.
     * @param {...string***REMOVED*** plugins The names of the plugins to load.
     * @returns {Object[]***REMOVED*** An array of flag-config objects representing the plugins.
     */
    plugins(...plugins) {
        return this.config({
            plugins
        ***REMOVED***);
    ***REMOVED***
***REMOVED***

export { FlatCompat ***REMOVED***;
