// eslint/lib/linter/code-path-analysis/code-path-segment.js
/**
 * @fileoverview The CodePathSegment class.
 * @author Toru Nagashima
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const debug = require("./debug-helpers");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Checks whether or not a given segment is reachable.
 * @param {CodePathSegment***REMOVED*** segment A segment to check.
 * @returns {boolean***REMOVED*** `true` if the segment is reachable.
 */
function isReachable(segment) {
    return segment.reachable;
***REMOVED***

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

/**
 * A code path segment.
 *
 * Each segment is arranged in a series of linked lists (implemented by arrays)
 * that keep track of the previous and next segments in a code path. In this way,
 * you can navigate between all segments in any code path so long as you have a
 * reference to any segment in that code path.
 *
 * When first created, the segment is in a detached state, meaning that it knows the
 * segments that came before it but those segments don't know that this new segment
 * follows it. Only when `CodePathSegment#markUsed()` is called on a segment does it
 * officially become part of the code path by updating the previous segments to know
 * that this new segment follows.
 */
class CodePathSegment {

    /**
     * Creates a new instance.
     * @param {string***REMOVED*** id An identifier.
     * @param {CodePathSegment[]***REMOVED*** allPrevSegments An array of the previous segments.
     *   This array includes unreachable segments.
     * @param {boolean***REMOVED*** reachable A flag which shows this is reachable.
     */
    constructor(id, allPrevSegments, reachable) {

        /**
         * The identifier of this code path.
         * Rules use it to store additional information of each rule.
         * @type {string***REMOVED***
         */
        this.id = id;

        /**
         * An array of the next reachable segments.
         * @type {CodePathSegment[]***REMOVED***
         */
        this.nextSegments = [];

        /**
         * An array of the previous reachable segments.
         * @type {CodePathSegment[]***REMOVED***
         */
        this.prevSegments = allPrevSegments.filter(isReachable);

        /**
         * An array of all next segments including reachable and unreachable.
         * @type {CodePathSegment[]***REMOVED***
         */
        this.allNextSegments = [];

        /**
         * An array of all previous segments including reachable and unreachable.
         * @type {CodePathSegment[]***REMOVED***
         */
        this.allPrevSegments = allPrevSegments;

        /**
         * A flag which shows this is reachable.
         * @type {boolean***REMOVED***
         */
        this.reachable = reachable;

        // Internal data.
        Object.defineProperty(this, "internal", {
            value: {

                // determines if the segment has been attached to the code path
                used: false,

                // array of previous segments coming from the end of a loop
                loopedPrevSegments: []
            ***REMOVED***
        ***REMOVED***);

        /* c8 ignore start */
        if (debug.enabled) {
            this.internal.nodes = [];
        ***REMOVED***/* c8 ignore stop */
    ***REMOVED***

    /**
     * Checks a given previous segment is coming from the end of a loop.
     * @param {CodePathSegment***REMOVED*** segment A previous segment to check.
     * @returns {boolean***REMOVED*** `true` if the segment is coming from the end of a loop.
     */
    isLoopedPrevSegment(segment) {
        return this.internal.loopedPrevSegments.includes(segment);
    ***REMOVED***

    /**
     * Creates the root segment.
     * @param {string***REMOVED*** id An identifier.
     * @returns {CodePathSegment***REMOVED*** The created segment.
     */
    static newRoot(id) {
        return new CodePathSegment(id, [], true);
    ***REMOVED***

    /**
     * Creates a new segment and appends it after the given segments.
     * @param {string***REMOVED*** id An identifier.
     * @param {CodePathSegment[]***REMOVED*** allPrevSegments An array of the previous segments
     *      to append to.
     * @returns {CodePathSegment***REMOVED*** The created segment.
     */
    static newNext(id, allPrevSegments) {
        return new CodePathSegment(
            id,
            CodePathSegment.flattenUnusedSegments(allPrevSegments),
            allPrevSegments.some(isReachable)
        );
    ***REMOVED***

    /**
     * Creates an unreachable segment and appends it after the given segments.
     * @param {string***REMOVED*** id An identifier.
     * @param {CodePathSegment[]***REMOVED*** allPrevSegments An array of the previous segments.
     * @returns {CodePathSegment***REMOVED*** The created segment.
     */
    static newUnreachable(id, allPrevSegments) {
        const segment = new CodePathSegment(id, CodePathSegment.flattenUnusedSegments(allPrevSegments), false);

        /*
         * In `if (a) return a; foo();` case, the unreachable segment preceded by
         * the return statement is not used but must not be removed.
         */
        CodePathSegment.markUsed(segment);

        return segment;
    ***REMOVED***

    /**
     * Creates a segment that follows given segments.
     * This factory method does not connect with `allPrevSegments`.
     * But this inherits `reachable` flag.
     * @param {string***REMOVED*** id An identifier.
     * @param {CodePathSegment[]***REMOVED*** allPrevSegments An array of the previous segments.
     * @returns {CodePathSegment***REMOVED*** The created segment.
     */
    static newDisconnected(id, allPrevSegments) {
        return new CodePathSegment(id, [], allPrevSegments.some(isReachable));
    ***REMOVED***

    /**
     * Marks a given segment as used.
     *
     * And this function registers the segment into the previous segments as a next.
     * @param {CodePathSegment***REMOVED*** segment A segment to mark.
     * @returns {void***REMOVED***
     */
    static markUsed(segment) {
        if (segment.internal.used) {
            return;
        ***REMOVED***
        segment.internal.used = true;

        let i;

        if (segment.reachable) {

            /*
             * If the segment is reachable, then it's officially part of the
             * code path. This loops through all previous segments to update
             * their list of next segments. Because the segment is reachable,
             * it's added to both `nextSegments` and `allNextSegments`.
             */
            for (i = 0; i < segment.allPrevSegments.length; ++i) {
                const prevSegment = segment.allPrevSegments[i];

                prevSegment.allNextSegments.push(segment);
                prevSegment.nextSegments.push(segment);
            ***REMOVED***
        ***REMOVED*** else {

            /*
             * If the segment is not reachable, then it's not officially part of the
             * code path. This loops through all previous segments to update
             * their list of next segments. Because the segment is not reachable,
             * it's added only to `allNextSegments`.
             */
            for (i = 0; i < segment.allPrevSegments.length; ++i) {
                segment.allPrevSegments[i].allNextSegments.push(segment);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    /**
     * Marks a previous segment as looped.
     * @param {CodePathSegment***REMOVED*** segment A segment.
     * @param {CodePathSegment***REMOVED*** prevSegment A previous segment to mark.
     * @returns {void***REMOVED***
     */
    static markPrevSegmentAsLooped(segment, prevSegment) {
        segment.internal.loopedPrevSegments.push(prevSegment);
    ***REMOVED***

    /**
     * Creates a new array based on an array of segments. If any segment in the
     * array is unused, then it is replaced by all of its previous segments.
     * All used segments are returned as-is without replacement.
     * @param {CodePathSegment[]***REMOVED*** segments The array of segments to flatten.
     * @returns {CodePathSegment[]***REMOVED*** The flattened array.
     */
    static flattenUnusedSegments(segments) {
        const done = new Set();

        for (let i = 0; i < segments.length; ++i) {
            const segment = segments[i];

            // Ignores duplicated.
            if (done.has(segment)) {
                continue;
            ***REMOVED***

            // Use previous segments if unused.
            if (!segment.internal.used) {
                for (let j = 0; j < segment.allPrevSegments.length; ++j) {
                    const prevSegment = segment.allPrevSegments[j];

                    if (!done.has(prevSegment)) {
                        done.add(prevSegment);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED*** else {
                done.add(segment);
            ***REMOVED***
        ***REMOVED***

        return [...done];
    ***REMOVED***
***REMOVED***

module.exports = CodePathSegment;
