// eslint/lib/linter/code-path-analysis/fork-context.js
/**
 * @fileoverview A class to operate forking.
 *
 * This is state of forking.
 * This has a fork list and manages it.
 *
 * @author Toru Nagashima
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const assert = require("assert"),
    CodePathSegment = require("./code-path-segment");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Gets whether or not a given segment is reachable.
 * @param {CodePathSegment***REMOVED*** segment A segment to get.
 * @returns {boolean***REMOVED*** `true` if the segment is reachable.
 */
function isReachable(segment) {
    return segment.reachable;
***REMOVED***

/**
 * Creates new segments from the specific range of `context.segmentsList`.
 *
 * When `context.segmentsList` is `[[a, b], [c, d], [e, f]]`, `begin` is `0`, and
 * `end` is `-1`, this creates `[g, h]`. This `g` is from `a`, `c`, and `e`.
 * This `h` is from `b`, `d`, and `f`.
 * @param {ForkContext***REMOVED*** context An instance.
 * @param {number***REMOVED*** begin The first index of the previous segments.
 * @param {number***REMOVED*** end The last index of the previous segments.
 * @param {Function***REMOVED*** create A factory function of new segments.
 * @returns {CodePathSegment[]***REMOVED*** New segments.
 */
function makeSegments(context, begin, end, create) {
    const list = context.segmentsList;

    const normalizedBegin = begin >= 0 ? begin : list.length + begin;
    const normalizedEnd = end >= 0 ? end : list.length + end;

    const segments = [];

    for (let i = 0; i < context.count; ++i) {
        const allPrevSegments = [];

        for (let j = normalizedBegin; j <= normalizedEnd; ++j) {
            allPrevSegments.push(list[j][i]);
        ***REMOVED***

        segments.push(create(context.idGenerator.next(), allPrevSegments));
    ***REMOVED***

    return segments;
***REMOVED***

/**
 * `segments` becomes doubly in a `finally` block. Then if a code path exits by a
 * control statement (such as `break`, `continue`) from the `finally` block, the
 * destination's segments may be half of the source segments. In that case, this
 * merges segments.
 * @param {ForkContext***REMOVED*** context An instance.
 * @param {CodePathSegment[]***REMOVED*** segments Segments to merge.
 * @returns {CodePathSegment[]***REMOVED*** The merged segments.
 */
function mergeExtraSegments(context, segments) {
    let currentSegments = segments;

    while (currentSegments.length > context.count) {
        const merged = [];

        for (let i = 0, length = currentSegments.length / 2 | 0; i < length; ++i) {
            merged.push(CodePathSegment.newNext(
                context.idGenerator.next(),
                [currentSegments[i], currentSegments[i + length]]
            ));
        ***REMOVED***
        currentSegments = merged;
    ***REMOVED***
    return currentSegments;
***REMOVED***

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

/**
 * A class to manage forking.
 */
class ForkContext {

    // eslint-disable-next-line jsdoc/require-description
    /**
     * @param {IdGenerator***REMOVED*** idGenerator An identifier generator for segments.
     * @param {ForkContext|null***REMOVED*** upper An upper fork context.
     * @param {number***REMOVED*** count A number of parallel segments.
     */
    constructor(idGenerator, upper, count) {
        this.idGenerator = idGenerator;
        this.upper = upper;
        this.count = count;
        this.segmentsList = [];
    ***REMOVED***

    /**
     * The head segments.
     * @type {CodePathSegment[]***REMOVED***
     */
    get head() {
        const list = this.segmentsList;

        return list.length === 0 ? [] : list[list.length - 1];
    ***REMOVED***

    /**
     * A flag which shows empty.
     * @type {boolean***REMOVED***
     */
    get empty() {
        return this.segmentsList.length === 0;
    ***REMOVED***

    /**
     * A flag which shows reachable.
     * @type {boolean***REMOVED***
     */
    get reachable() {
        const segments = this.head;

        return segments.length > 0 && segments.some(isReachable);
    ***REMOVED***

    /**
     * Creates new segments from this context.
     * @param {number***REMOVED*** begin The first index of previous segments.
     * @param {number***REMOVED*** end The last index of previous segments.
     * @returns {CodePathSegment[]***REMOVED*** New segments.
     */
    makeNext(begin, end) {
        return makeSegments(this, begin, end, CodePathSegment.newNext);
    ***REMOVED***

    /**
     * Creates new segments from this context.
     * The new segments is always unreachable.
     * @param {number***REMOVED*** begin The first index of previous segments.
     * @param {number***REMOVED*** end The last index of previous segments.
     * @returns {CodePathSegment[]***REMOVED*** New segments.
     */
    makeUnreachable(begin, end) {
        return makeSegments(this, begin, end, CodePathSegment.newUnreachable);
    ***REMOVED***

    /**
     * Creates new segments from this context.
     * The new segments don't have connections for previous segments.
     * But these inherit the reachable flag from this context.
     * @param {number***REMOVED*** begin The first index of previous segments.
     * @param {number***REMOVED*** end The last index of previous segments.
     * @returns {CodePathSegment[]***REMOVED*** New segments.
     */
    makeDisconnected(begin, end) {
        return makeSegments(this, begin, end, CodePathSegment.newDisconnected);
    ***REMOVED***

    /**
     * Adds segments into this context.
     * The added segments become the head.
     * @param {CodePathSegment[]***REMOVED*** segments Segments to add.
     * @returns {void***REMOVED***
     */
    add(segments) {
        assert(segments.length >= this.count, `${segments.length***REMOVED*** >= ${this.count***REMOVED***`);

        this.segmentsList.push(mergeExtraSegments(this, segments));
    ***REMOVED***

    /**
     * Replaces the head segments with given segments.
     * The current head segments are removed.
     * @param {CodePathSegment[]***REMOVED*** segments Segments to add.
     * @returns {void***REMOVED***
     */
    replaceHead(segments) {
        assert(segments.length >= this.count, `${segments.length***REMOVED*** >= ${this.count***REMOVED***`);

        this.segmentsList.splice(-1, 1, mergeExtraSegments(this, segments));
    ***REMOVED***

    /**
     * Adds all segments of a given fork context into this context.
     * @param {ForkContext***REMOVED*** context A fork context to add.
     * @returns {void***REMOVED***
     */
    addAll(context) {
        assert(context.count === this.count);

        const source = context.segmentsList;

        for (let i = 0; i < source.length; ++i) {
            this.segmentsList.push(source[i]);
        ***REMOVED***
    ***REMOVED***

    /**
     * Clears all segments in this context.
     * @returns {void***REMOVED***
     */
    clear() {
        this.segmentsList = [];
    ***REMOVED***

    /**
     * Creates the root fork context.
     * @param {IdGenerator***REMOVED*** idGenerator An identifier generator for segments.
     * @returns {ForkContext***REMOVED*** New fork context.
     */
    static newRoot(idGenerator) {
        const context = new ForkContext(idGenerator, null, 1);

        context.add([CodePathSegment.newRoot(idGenerator.next())]);

        return context;
    ***REMOVED***

    /**
     * Creates an empty fork context preceded by a given context.
     * @param {ForkContext***REMOVED*** parentContext The parent fork context.
     * @param {boolean***REMOVED*** forkLeavingPath A flag which shows inside of `finally` block.
     * @returns {ForkContext***REMOVED*** New fork context.
     */
    static newEmpty(parentContext, forkLeavingPath) {
        return new ForkContext(
            parentContext.idGenerator,
            parentContext,
            (forkLeavingPath ? 2 : 1) * parentContext.count
        );
    ***REMOVED***
***REMOVED***

module.exports = ForkContext;
