// @nodelib/fs.walk/out/readers/async.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const events_1 = require("events");
const fsScandir = require("@nodelib/fs.scandir");
const fastq = require("fastq");
const common = require("./common");
const reader_1 = require("./reader");
class AsyncReader extends reader_1.default {
    constructor(_root, _settings) {
        super(_root, _settings);
        this._settings = _settings;
        this._scandir = fsScandir.scandir;
        this._emitter = new events_1.EventEmitter();
        this._queue = fastq(this._worker.bind(this), this._settings.concurrency);
        this._isFatalError = false;
        this._isDestroyed = false;
        this._queue.drain = () => {
            if (!this._isFatalError) {
                this._emitter.emit('end');
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
    read() {
        this._isFatalError = false;
        this._isDestroyed = false;
        setImmediate(() => {
            this._pushToQueue(this._root, this._settings.basePath);
        ***REMOVED***);
        return this._emitter;
    ***REMOVED***
    get isDestroyed() {
        return this._isDestroyed;
    ***REMOVED***
    destroy() {
        if (this._isDestroyed) {
            throw new Error('The reader is already destroyed');
        ***REMOVED***
        this._isDestroyed = true;
        this._queue.killAndDrain();
    ***REMOVED***
    onEntry(callback) {
        this._emitter.on('entry', callback);
    ***REMOVED***
    onError(callback) {
        this._emitter.once('error', callback);
    ***REMOVED***
    onEnd(callback) {
        this._emitter.once('end', callback);
    ***REMOVED***
    _pushToQueue(directory, base) {
        const queueItem = { directory, base ***REMOVED***;
        this._queue.push(queueItem, (error) => {
            if (error !== null) {
                this._handleError(error);
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***
    _worker(item, done) {
        this._scandir(item.directory, this._settings.fsScandirSettings, (error, entries) => {
            if (error !== null) {
                done(error, undefined);
                return;
            ***REMOVED***
            for (const entry of entries) {
                this._handleEntry(entry, item.base);
            ***REMOVED***
            done(null, undefined);
        ***REMOVED***);
    ***REMOVED***
    _handleError(error) {
        if (this._isDestroyed || !common.isFatalError(this._settings, error)) {
            return;
        ***REMOVED***
        this._isFatalError = true;
        this._isDestroyed = true;
        this._emitter.emit('error', error);
    ***REMOVED***
    _handleEntry(entry, base) {
        if (this._isDestroyed || this._isFatalError) {
            return;
        ***REMOVED***
        const fullpath = entry.path;
        if (base !== undefined) {
            entry.path = common.joinPathSegments(base, entry.name, this._settings.pathSegmentSeparator);
        ***REMOVED***
        if (common.isAppliedFilter(this._settings.entryFilter, entry)) {
            this._emitEntry(entry);
        ***REMOVED***
        if (entry.dirent.isDirectory() && common.isAppliedFilter(this._settings.deepFilter, entry)) {
            this._pushToQueue(fullpath, base === undefined ? undefined : entry.path);
        ***REMOVED***
    ***REMOVED***
    _emitEntry(entry) {
        this._emitter.emit('entry', entry);
    ***REMOVED***
***REMOVED***
exports.default = AsyncReader;
