// webpack-sources/lib/helpers/getFromStreamChunks.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const createMappingsSerializer = require("./createMappingsSerializer");

exports.getSourceAndMap = (inputSource, options) => {
	let code = "";
	let mappings = "";
	let sources = [];
	let sourcesContent = [];
	let names = [];
	const addMapping = createMappingsSerializer(options);
	const { source ***REMOVED*** = inputSource.streamChunks(
		Object.assign({***REMOVED***, options, { finalSource: true ***REMOVED***),
		(
			chunk,
			generatedLine,
			generatedColumn,
			sourceIndex,
			originalLine,
			originalColumn,
			nameIndex
		) => {
			if (chunk !== undefined) code += chunk;
			mappings += addMapping(
				generatedLine,
				generatedColumn,
				sourceIndex,
				originalLine,
				originalColumn,
				nameIndex
			);
		***REMOVED***,
		(sourceIndex, source, sourceContent) => {
			while (sources.length < sourceIndex) {
				sources.push(null);
			***REMOVED***
			sources[sourceIndex] = source;
			if (sourceContent !== undefined) {
				while (sourcesContent.length < sourceIndex) {
					sourcesContent.push(null);
				***REMOVED***
				sourcesContent[sourceIndex] = sourceContent;
			***REMOVED***
		***REMOVED***,
		(nameIndex, name) => {
			while (names.length < nameIndex) {
				names.push(null);
			***REMOVED***
			names[nameIndex] = name;
		***REMOVED***
	);
	return {
		source: source !== undefined ? source : code,
		map:
			mappings.length > 0
				? {
						version: 3,
						file: "x",
						mappings,
						sources,
						sourcesContent:
							sourcesContent.length > 0 ? sourcesContent : undefined,
						names
				  ***REMOVED***
				: null
	***REMOVED***;
***REMOVED***;

exports.getMap = (source, options) => {
	let mappings = "";
	let sources = [];
	let sourcesContent = [];
	let names = [];
	const addMapping = createMappingsSerializer(options);
	source.streamChunks(
		Object.assign({***REMOVED***, options, { source: false, finalSource: true ***REMOVED***),
		(
			chunk,
			generatedLine,
			generatedColumn,
			sourceIndex,
			originalLine,
			originalColumn,
			nameIndex
		) => {
			mappings += addMapping(
				generatedLine,
				generatedColumn,
				sourceIndex,
				originalLine,
				originalColumn,
				nameIndex
			);
		***REMOVED***,
		(sourceIndex, source, sourceContent) => {
			while (sources.length < sourceIndex) {
				sources.push(null);
			***REMOVED***
			sources[sourceIndex] = source;
			if (sourceContent !== undefined) {
				while (sourcesContent.length < sourceIndex) {
					sourcesContent.push(null);
				***REMOVED***
				sourcesContent[sourceIndex] = sourceContent;
			***REMOVED***
		***REMOVED***,
		(nameIndex, name) => {
			while (names.length < nameIndex) {
				names.push(null);
			***REMOVED***
			names[nameIndex] = name;
		***REMOVED***
	);
	return mappings.length > 0
		? {
				version: 3,
				file: "x",
				mappings,
				sources,
				sourcesContent: sourcesContent.length > 0 ? sourcesContent : undefined,
				names
		  ***REMOVED***
		: null;
***REMOVED***;
