// keyv/src/index.js
'use strict';

const EventEmitter = require('events');
const JSONB = require('json-buffer');

const loadStore = options => {
	const adapters = {
		redis: '@keyv/redis',
		rediss: '@keyv/redis',
		mongodb: '@keyv/mongo',
		mongo: '@keyv/mongo',
		sqlite: '@keyv/sqlite',
		postgresql: '@keyv/postgres',
		postgres: '@keyv/postgres',
		mysql: '@keyv/mysql',
		etcd: '@keyv/etcd',
		offline: '@keyv/offline',
		tiered: '@keyv/tiered',
	***REMOVED***;
	if (options.adapter || options.uri) {
		const adapter = options.adapter || /^[^:+]*/.exec(options.uri)[0];
		return new (require(adapters[adapter]))(options);
	***REMOVED***

	return new Map();
***REMOVED***;

const iterableAdapters = [
	'sqlite',
	'postgres',
	'mysql',
	'mongo',
	'redis',
	'tiered',
];

class Keyv extends EventEmitter {
	constructor(uri, {emitErrors = true, ...options***REMOVED*** = {***REMOVED***) {
		super();
		this.opts = {
			namespace: 'keyv',
			serialize: JSONB.stringify,
			deserialize: JSONB.parse,
			...((typeof uri === 'string') ? {uri***REMOVED*** : uri),
			...options,
		***REMOVED***;

		if (!this.opts.store) {
			const adapterOptions = {...this.opts***REMOVED***;
			this.opts.store = loadStore(adapterOptions);
		***REMOVED***

		if (this.opts.compression) {
			const compression = this.opts.compression;
			this.opts.serialize = compression.serialize.bind(compression);
			this.opts.deserialize = compression.deserialize.bind(compression);
		***REMOVED***

		if (typeof this.opts.store.on === 'function' && emitErrors) {
			this.opts.store.on('error', error => this.emit('error', error));
		***REMOVED***

		this.opts.store.namespace = this.opts.namespace;

		const generateIterator = iterator => async function * () {
			for await (const [key, raw] of typeof iterator === 'function'
				? iterator(this.opts.store.namespace)
				: iterator) {
				const data = await this.opts.deserialize(raw);
				if (this.opts.store.namespace && !key.includes(this.opts.store.namespace)) {
					continue;
				***REMOVED***

				if (typeof data.expires === 'number' && Date.now() > data.expires) {
					this.delete(key);
					continue;
				***REMOVED***

				yield [this._getKeyUnprefix(key), data.value];
			***REMOVED***
		***REMOVED***;

		// Attach iterators
		if (typeof this.opts.store[Symbol.iterator] === 'function' && this.opts.store instanceof Map) {
			this.iterator = generateIterator(this.opts.store);
		***REMOVED*** else if (typeof this.opts.store.iterator === 'function' && this.opts.store.opts
			&& this._checkIterableAdaptar()) {
			this.iterator = generateIterator(this.opts.store.iterator.bind(this.opts.store));
		***REMOVED***
	***REMOVED***

	_checkIterableAdaptar() {
		return iterableAdapters.includes(this.opts.store.opts.dialect)
			|| iterableAdapters.findIndex(element => this.opts.store.opts.url.includes(element)) >= 0;
	***REMOVED***

	_getKeyPrefix(key) {
		return `${this.opts.namespace***REMOVED***:${key***REMOVED***`;
	***REMOVED***

	_getKeyPrefixArray(keys) {
		return keys.map(key => `${this.opts.namespace***REMOVED***:${key***REMOVED***`);
	***REMOVED***

	_getKeyUnprefix(key) {
		return key
			.split(':')
			.splice(1)
			.join(':');
	***REMOVED***

	get(key, options) {
		const {store***REMOVED*** = this.opts;
		const isArray = Array.isArray(key);
		const keyPrefixed = isArray ? this._getKeyPrefixArray(key) : this._getKeyPrefix(key);
		if (isArray && store.getMany === undefined) {
			const promises = [];
			for (const key of keyPrefixed) {
				promises.push(Promise.resolve()
					.then(() => store.get(key))
					.then(data => (typeof data === 'string') ? this.opts.deserialize(data) : (this.opts.compression ? this.opts.deserialize(data) : data))
					.then(data => {
						if (data === undefined || data === null) {
							return undefined;
						***REMOVED***

						if (typeof data.expires === 'number' && Date.now() > data.expires) {
							return this.delete(key).then(() => undefined);
						***REMOVED***

						return (options && options.raw) ? data : data.value;
					***REMOVED***),
				);
			***REMOVED***

			return Promise.allSettled(promises)
				.then(values => {
					const data = [];
					for (const value of values) {
						data.push(value.value);
					***REMOVED***

					return data;
				***REMOVED***);
		***REMOVED***

		return Promise.resolve()
			.then(() => isArray ? store.getMany(keyPrefixed) : store.get(keyPrefixed))
			.then(data => (typeof data === 'string') ? this.opts.deserialize(data) : (this.opts.compression ? this.opts.deserialize(data) : data))
			.then(data => {
				if (data === undefined || data === null) {
					return undefined;
				***REMOVED***

				if (isArray) {
					return data.map((row, index) => {
						if ((typeof row === 'string')) {
							row = this.opts.deserialize(row);
						***REMOVED***

						if (row === undefined || row === null) {
							return undefined;
						***REMOVED***

						if (typeof row.expires === 'number' && Date.now() > row.expires) {
							this.delete(key[index]).then(() => undefined);
							return undefined;
						***REMOVED***

						return (options && options.raw) ? row : row.value;
					***REMOVED***);
				***REMOVED***

				if (typeof data.expires === 'number' && Date.now() > data.expires) {
					return this.delete(key).then(() => undefined);
				***REMOVED***

				return (options && options.raw) ? data : data.value;
			***REMOVED***);
	***REMOVED***

	set(key, value, ttl) {
		const keyPrefixed = this._getKeyPrefix(key);
		if (typeof ttl === 'undefined') {
			ttl = this.opts.ttl;
		***REMOVED***

		if (ttl === 0) {
			ttl = undefined;
		***REMOVED***

		const {store***REMOVED*** = this.opts;

		return Promise.resolve()
			.then(() => {
				const expires = (typeof ttl === 'number') ? (Date.now() + ttl) : null;
				if (typeof value === 'symbol') {
					this.emit('error', 'symbol cannot be serialized');
				***REMOVED***

				value = {value, expires***REMOVED***;
				return this.opts.serialize(value);
			***REMOVED***)
			.then(value => store.set(keyPrefixed, value, ttl))
			.then(() => true);
	***REMOVED***

	delete(key) {
		const {store***REMOVED*** = this.opts;
		if (Array.isArray(key)) {
			const keyPrefixed = this._getKeyPrefixArray(key);
			if (store.deleteMany === undefined) {
				const promises = [];
				for (const key of keyPrefixed) {
					promises.push(store.delete(key));
				***REMOVED***

				return Promise.allSettled(promises)
					.then(values => values.every(x => x.value === true));
			***REMOVED***

			return Promise.resolve()
				.then(() => store.deleteMany(keyPrefixed));
		***REMOVED***

		const keyPrefixed = this._getKeyPrefix(key);
		return Promise.resolve()
			.then(() => store.delete(keyPrefixed));
	***REMOVED***

	clear() {
		const {store***REMOVED*** = this.opts;
		return Promise.resolve()
			.then(() => store.clear());
	***REMOVED***

	has(key) {
		const keyPrefixed = this._getKeyPrefix(key);
		const {store***REMOVED*** = this.opts;
		return Promise.resolve()
			.then(async () => {
				if (typeof store.has === 'function') {
					return store.has(keyPrefixed);
				***REMOVED***

				const value = await store.get(keyPrefixed);
				return value !== undefined;
			***REMOVED***);
	***REMOVED***

	disconnect() {
		const {store***REMOVED*** = this.opts;
		if (typeof store.disconnect === 'function') {
			return store.disconnect();
		***REMOVED***
	***REMOVED***
***REMOVED***

module.exports = Keyv;
