// promise/src/core.js
'use strict';

var asap = require('asap/raw');

function noop() {***REMOVED***

// States:
//
// 0 - pending
// 1 - fulfilled with _value
// 2 - rejected with _value
// 3 - adopted the state of another promise, _value
//
// once the state is no longer pending (0) it is immutable

// All `_` prefixed properties will be reduced to `_{random number***REMOVED***`
// at build time to obfuscate them and discourage their use.
// We don't use symbols or Object.defineProperty to fully hide them
// because the performance isn't good enough.


// to avoid using try/catch inside critical functions, we
// extract them to here.
var LAST_ERROR = null;
var IS_ERROR = {***REMOVED***;
function getThen(obj) {
  try {
    return obj.then;
  ***REMOVED*** catch (ex) {
    LAST_ERROR = ex;
    return IS_ERROR;
  ***REMOVED***
***REMOVED***

function tryCallOne(fn, a) {
  try {
    return fn(a);
  ***REMOVED*** catch (ex) {
    LAST_ERROR = ex;
    return IS_ERROR;
  ***REMOVED***
***REMOVED***
function tryCallTwo(fn, a, b) {
  try {
    fn(a, b);
  ***REMOVED*** catch (ex) {
    LAST_ERROR = ex;
    return IS_ERROR;
  ***REMOVED***
***REMOVED***

module.exports = Promise;

function Promise(fn) {
  if (typeof this !== 'object') {
    throw new TypeError('Promises must be constructed via new');
  ***REMOVED***
  if (typeof fn !== 'function') {
    throw new TypeError('Promise constructor\'s argument is not a function');
  ***REMOVED***
  this._deferredState = 0;
  this._state = 0;
  this._value = null;
  this._deferreds = null;
  if (fn === noop) return;
  doResolve(fn, this);
***REMOVED***
Promise._onHandle = null;
Promise._onReject = null;
Promise._noop = noop;

Promise.prototype.then = function(onFulfilled, onRejected) {
  if (this.constructor !== Promise) {
    return safeThen(this, onFulfilled, onRejected);
  ***REMOVED***
  var res = new Promise(noop);
  handle(this, new Handler(onFulfilled, onRejected, res));
  return res;
***REMOVED***;

function safeThen(self, onFulfilled, onRejected) {
  return new self.constructor(function (resolve, reject) {
    var res = new Promise(noop);
    res.then(resolve, reject);
    handle(self, new Handler(onFulfilled, onRejected, res));
  ***REMOVED***);
***REMOVED***
function handle(self, deferred) {
  while (self._state === 3) {
    self = self._value;
  ***REMOVED***
  if (Promise._onHandle) {
    Promise._onHandle(self);
  ***REMOVED***
  if (self._state === 0) {
    if (self._deferredState === 0) {
      self._deferredState = 1;
      self._deferreds = deferred;
      return;
    ***REMOVED***
    if (self._deferredState === 1) {
      self._deferredState = 2;
      self._deferreds = [self._deferreds, deferred];
      return;
    ***REMOVED***
    self._deferreds.push(deferred);
    return;
  ***REMOVED***
  handleResolved(self, deferred);
***REMOVED***

function handleResolved(self, deferred) {
  asap(function() {
    var cb = self._state === 1 ? deferred.onFulfilled : deferred.onRejected;
    if (cb === null) {
      if (self._state === 1) {
        resolve(deferred.promise, self._value);
      ***REMOVED*** else {
        reject(deferred.promise, self._value);
      ***REMOVED***
      return;
    ***REMOVED***
    var ret = tryCallOne(cb, self._value);
    if (ret === IS_ERROR) {
      reject(deferred.promise, LAST_ERROR);
    ***REMOVED*** else {
      resolve(deferred.promise, ret);
    ***REMOVED***
  ***REMOVED***);
***REMOVED***
function resolve(self, newValue) {
  // Promise Resolution Procedure: https://github.com/promises-aplus/promises-spec#the-promise-resolution-procedure
  if (newValue === self) {
    return reject(
      self,
      new TypeError('A promise cannot be resolved with itself.')
    );
  ***REMOVED***
  if (
    newValue &&
    (typeof newValue === 'object' || typeof newValue === 'function')
  ) {
    var then = getThen(newValue);
    if (then === IS_ERROR) {
      return reject(self, LAST_ERROR);
    ***REMOVED***
    if (
      then === self.then &&
      newValue instanceof Promise
    ) {
      self._state = 3;
      self._value = newValue;
      finale(self);
      return;
    ***REMOVED*** else if (typeof then === 'function') {
      doResolve(then.bind(newValue), self);
      return;
    ***REMOVED***
  ***REMOVED***
  self._state = 1;
  self._value = newValue;
  finale(self);
***REMOVED***

function reject(self, newValue) {
  self._state = 2;
  self._value = newValue;
  if (Promise._onReject) {
    Promise._onReject(self, newValue);
  ***REMOVED***
  finale(self);
***REMOVED***
function finale(self) {
  if (self._deferredState === 1) {
    handle(self, self._deferreds);
    self._deferreds = null;
  ***REMOVED***
  if (self._deferredState === 2) {
    for (var i = 0; i < self._deferreds.length; i++) {
      handle(self, self._deferreds[i]);
    ***REMOVED***
    self._deferreds = null;
  ***REMOVED***
***REMOVED***

function Handler(onFulfilled, onRejected, promise){
  this.onFulfilled = typeof onFulfilled === 'function' ? onFulfilled : null;
  this.onRejected = typeof onRejected === 'function' ? onRejected : null;
  this.promise = promise;
***REMOVED***

/**
 * Take a potentially misbehaving resolver function and make sure
 * onFulfilled and onRejected are only called once.
 *
 * Makes no guarantees about asynchrony.
 */
function doResolve(fn, promise) {
  var done = false;
  var res = tryCallTwo(fn, function (value) {
    if (done) return;
    done = true;
    resolve(promise, value);
  ***REMOVED***, function (reason) {
    if (done) return;
    done = true;
    reject(promise, reason);
  ***REMOVED***);
  if (!done && res === IS_ERROR) {
    done = true;
    reject(promise, LAST_ERROR);
  ***REMOVED***
***REMOVED***
