// promise/src/synchronous.js
'use strict';

var Promise = require('./core.js');

module.exports = Promise;
Promise.enableSynchronous = function () {
  Promise.prototype.isPending = function() {
    return this.getState() == 0;
  ***REMOVED***;

  Promise.prototype.isFulfilled = function() {
    return this.getState() == 1;
  ***REMOVED***;

  Promise.prototype.isRejected = function() {
    return this.getState() == 2;
  ***REMOVED***;

  Promise.prototype.getValue = function () {
    if (this._state === 3) {
      return this._value.getValue();
    ***REMOVED***

    if (!this.isFulfilled()) {
      throw new Error('Cannot get a value of an unfulfilled promise.');
    ***REMOVED***

    return this._value;
  ***REMOVED***;

  Promise.prototype.getReason = function () {
    if (this._state === 3) {
      return this._value.getReason();
    ***REMOVED***

    if (!this.isRejected()) {
      throw new Error('Cannot get a rejection reason of a non-rejected promise.');
    ***REMOVED***

    return this._value;
  ***REMOVED***;

  Promise.prototype.getState = function () {
    if (this._state === 3) {
      return this._value.getState();
    ***REMOVED***
    if (this._state === -1 || this._state === -2) {
      return 0;
    ***REMOVED***

    return this._state;
  ***REMOVED***;
***REMOVED***;

Promise.disableSynchronous = function() {
  Promise.prototype.isPending = undefined;
  Promise.prototype.isFulfilled = undefined;
  Promise.prototype.isRejected = undefined;
  Promise.prototype.getValue = undefined;
  Promise.prototype.getReason = undefined;
  Promise.prototype.getState = undefined;
***REMOVED***;
