// yaml/browser/dist/schema/json/schema.js
import { Scalar ***REMOVED*** from '../../nodes/Scalar.js';
import { map ***REMOVED*** from '../common/map.js';
import { seq ***REMOVED*** from '../common/seq.js';

function intIdentify(value) {
    return typeof value === 'bigint' || Number.isInteger(value);
***REMOVED***
const stringifyJSON = ({ value ***REMOVED***) => JSON.stringify(value);
const jsonScalars = [
    {
        identify: value => typeof value === 'string',
        default: true,
        tag: 'tag:yaml.org,2002:str',
        resolve: str => str,
        stringify: stringifyJSON
    ***REMOVED***,
    {
        identify: value => value == null,
        createNode: () => new Scalar(null),
        default: true,
        tag: 'tag:yaml.org,2002:null',
        test: /^null$/,
        resolve: () => null,
        stringify: stringifyJSON
    ***REMOVED***,
    {
        identify: value => typeof value === 'boolean',
        default: true,
        tag: 'tag:yaml.org,2002:bool',
        test: /^true$|^false$/,
        resolve: str => str === 'true',
        stringify: stringifyJSON
    ***REMOVED***,
    {
        identify: intIdentify,
        default: true,
        tag: 'tag:yaml.org,2002:int',
        test: /^-?(?:0|[1-9][0-9]*)$/,
        resolve: (str, _onError, { intAsBigInt ***REMOVED***) => intAsBigInt ? BigInt(str) : parseInt(str, 10),
        stringify: ({ value ***REMOVED***) => intIdentify(value) ? value.toString() : JSON.stringify(value)
    ***REMOVED***,
    {
        identify: value => typeof value === 'number',
        default: true,
        tag: 'tag:yaml.org,2002:float',
        test: /^-?(?:0|[1-9][0-9]*)(?:\.[0-9]*)?(?:[eE][-+]?[0-9]+)?$/,
        resolve: str => parseFloat(str),
        stringify: stringifyJSON
    ***REMOVED***
];
const jsonError = {
    default: true,
    tag: '',
    test: /^/,
    resolve(str, onError) {
        onError(`Unresolved plain scalar ${JSON.stringify(str)***REMOVED***`);
        return str;
    ***REMOVED***
***REMOVED***;
const schema = [map, seq].concat(jsonScalars, jsonError);

export { schema ***REMOVED***;
