// tough-cookie/lib/utilHelper.js
function requireUtil() {
  try {
    // eslint-disable-next-line no-restricted-modules
    return require("util");
  ***REMOVED*** catch (e) {
    return null;
  ***REMOVED***
***REMOVED***

// for v10.12.0+
function lookupCustomInspectSymbol() {
  return Symbol.for("nodejs.util.inspect.custom");
***REMOVED***

// for older node environments
function tryReadingCustomSymbolFromUtilInspect(options) {
  const _requireUtil = options.requireUtil || requireUtil;
  const util = _requireUtil();
  return util ? util.inspect.custom : null;
***REMOVED***

exports.getUtilInspect = function getUtilInspect(fallback, options = {***REMOVED***) {
  const _requireUtil = options.requireUtil || requireUtil;
  const util = _requireUtil();
  return function inspect(value, showHidden, depth) {
    return util ? util.inspect(value, showHidden, depth) : fallback(value);
  ***REMOVED***;
***REMOVED***;

exports.getCustomInspectSymbol = function getCustomInspectSymbol(options = {***REMOVED***) {
  const _lookupCustomInspectSymbol =
    options.lookupCustomInspectSymbol || lookupCustomInspectSymbol;

  // get custom inspect symbol for node environments
  return (
    _lookupCustomInspectSymbol() ||
    tryReadingCustomSymbolFromUtilInspect(options)
  );
***REMOVED***;
