// jsdom/lib/jsdom/living/selection/Selection-impl.js
"use strict";

const DOMException = require("domexception/webidl2js-wrapper");

const NODE_TYPE = require("../node-type");
const { nodeLength, nodeRoot ***REMOVED*** = require("../helpers/node");
const { domSymbolTree ***REMOVED*** = require("../helpers/internal-constants");
const { compareBoundaryPointsPosition ***REMOVED*** = require("../range/boundary-point");

const { setBoundaryPointStart, setBoundaryPointEnd ***REMOVED*** = require("../range/Range-impl");

const Range = require("../generated/Range");
const { implForWrapper ***REMOVED*** = require("../generated/utils");

// https://w3c.github.io/selection-api/#dfn-direction
const SELECTION_DIRECTION = {
  FORWARDS: 1,
  BACKWARDS: -1,
  DIRECTIONLESS: 0
***REMOVED***;

// https://w3c.github.io/selection-api/#dom-selection
class SelectionImpl {
  constructor(globalObject) {
    this._range = null;
    this._direction = SELECTION_DIRECTION.DIRECTIONLESS;

    this._globalObject = globalObject;
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-anchornode
  get anchorNode() {
    const anchor = this._anchor;
    return anchor ? anchor.node : null;
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-anchoroffset
  get anchorOffset() {
    const anchor = this._anchor;
    return anchor ? anchor.offset : 0;
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-focusnode
  get focusNode() {
    const focus = this._focus;
    return focus ? focus.node : null;
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-focusoffset
  get focusOffset() {
    const focus = this._focus;
    return focus ? focus.offset : 0;
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-iscollapsed
  get isCollapsed() {
    return this._range === null || this._range.collapsed;
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-rangecount
  get rangeCount() {
    return this._isEmpty() ? 0 : 1;
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-type
  get type() {
    if (this._isEmpty()) {
      return "None";
    ***REMOVED*** else if (this._range.collapsed) {
      return "Caret";
    ***REMOVED***

    return "Range";
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-getrangeat
  getRangeAt(index) {
    if (index !== 0 || this._isEmpty()) {
      throw DOMException.create(this._globalObject, ["Invalid range index.", "IndexSizeError"]);
    ***REMOVED***

    return this._range;
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-addrange
  addRange(range) {
    if (range._root === implForWrapper(this._globalObject._document) && this.rangeCount === 0) {
      this._associateRange(range);
    ***REMOVED***
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-removerange
  removeRange(range) {
    if (range !== this._range) {
      throw DOMException.create(this._globalObject, ["Invalid range.", "NotFoundError"]);
    ***REMOVED***

    this._associateRange(null);
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-removeallranges
  removeAllRanges() {
    this._associateRange(null);
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-empty
  empty() {
    this.removeAllRanges();
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-collapse
  collapse(node, offset) {
    if (node === null) {
      this.removeAllRanges();
      return;
    ***REMOVED***

    if (node.nodeType === NODE_TYPE.DOCUMENT_TYPE_NODE) {
      throw DOMException.create(this._globalObject, [
        "DocumentType Node can't be used as boundary point.",
        "InvalidNodeTypeError"
      ]);
    ***REMOVED***

    if (offset > nodeLength(node)) {
      throw DOMException.create(this._globalObject, ["Invalid range index.", "IndexSizeError"]);
    ***REMOVED***

    if (nodeRoot(node) !== implForWrapper(this._globalObject._document)) {
      return;
    ***REMOVED***

    const newRange = Range.createImpl(this._globalObject, [], {
      start: { node, offset: 0 ***REMOVED***,
      end: { node, offset: 0 ***REMOVED***
    ***REMOVED***);

    setBoundaryPointStart(newRange, node, offset);
    setBoundaryPointEnd(newRange, node, offset);

    this._associateRange(newRange);
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-setposition
  setPosition(node, offset) {
    this.collapse(node, offset);
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-collapsetostart
  collapseToStart() {
    if (this._isEmpty()) {
      throw DOMException.create(this._globalObject, ["There is no selection to collapse.", "InvalidStateError"]);
    ***REMOVED***

    const { node, offset ***REMOVED*** = this._range._start;
    const newRange = Range.createImpl(this._globalObject, [], {
      start: { node, offset ***REMOVED***,
      end: { node, offset ***REMOVED***
    ***REMOVED***);

    this._associateRange(newRange);
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-collapsetoend
  collapseToEnd() {
    if (this._isEmpty()) {
      throw DOMException.create(this._globalObject, ["There is no selection to collapse.", "InvalidStateError"]);
    ***REMOVED***

    const { node, offset ***REMOVED*** = this._range._end;
    const newRange = Range.createImpl(this._globalObject, [], {
      start: { node, offset ***REMOVED***,
      end: { node, offset ***REMOVED***
    ***REMOVED***);

    this._associateRange(newRange);
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-extend
  extend(node, offset) {
    if (nodeRoot(node) !== implForWrapper(this._globalObject._document)) {
      return;
    ***REMOVED***

    if (this._isEmpty()) {
      throw DOMException.create(this._globalObject, ["There is no selection to extend.", "InvalidStateError"]);
    ***REMOVED***

    const { _anchor: oldAnchor ***REMOVED*** = this;
    const newFocus = { node, offset ***REMOVED***;

    const newRange = Range.createImpl(this._globalObject, [], {
      start: { node, offset: 0 ***REMOVED***,
      end: { node, offset: 0 ***REMOVED***
    ***REMOVED***);

    if (nodeRoot(node) !== this._range._root) {
      setBoundaryPointStart(newRange, newFocus.node, newFocus.offset);
      setBoundaryPointEnd(newRange, newFocus.node, newFocus.offset);
    ***REMOVED*** else if (compareBoundaryPointsPosition(oldAnchor, newFocus) <= 0) {
      setBoundaryPointStart(newRange, oldAnchor.node, oldAnchor.offset);
      setBoundaryPointEnd(newRange, newFocus.node, newFocus.offset);
    ***REMOVED*** else {
      setBoundaryPointStart(newRange, newFocus.node, newFocus.offset);
      setBoundaryPointEnd(newRange, oldAnchor.node, oldAnchor.offset);
    ***REMOVED***

    this._associateRange(newRange);

    this._direction = compareBoundaryPointsPosition(newFocus, oldAnchor) === -1 ?
      SELECTION_DIRECTION.BACKWARDS :
      SELECTION_DIRECTION.FORWARDS;
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-setbaseandextent
  setBaseAndExtent(anchorNode, anchorOffset, focusNode, focusOffset) {
    if (anchorOffset > nodeLength(anchorNode) || focusOffset > nodeLength(focusNode)) {
      throw DOMException.create(this._globalObject, ["Invalid anchor or focus offset.", "IndexSizeError"]);
    ***REMOVED***

    const document = implForWrapper(this._globalObject._document);
    if (document !== nodeRoot(anchorNode) || document !== nodeRoot(focusNode)) {
      return;
    ***REMOVED***

    const anchor = { node: anchorNode, offset: anchorOffset ***REMOVED***;
    const focus = { node: focusNode, offset: focusOffset ***REMOVED***;

    let newRange;
    if (compareBoundaryPointsPosition(anchor, focus) === -1) {
      newRange = Range.createImpl(this._globalObject, [], {
        start: { node: anchor.node, offset: anchor.offset ***REMOVED***,
        end: { node: focus.node, offset: focus.offset ***REMOVED***
      ***REMOVED***);
    ***REMOVED*** else {
      newRange = Range.createImpl(this._globalObject, [], {
        start: { node: focus.node, offset: focus.offset ***REMOVED***,
        end: { node: anchor.node, offset: anchor.offset ***REMOVED***
      ***REMOVED***);
    ***REMOVED***

    this._associateRange(newRange);

    this._direction = compareBoundaryPointsPosition(focus, anchor) === -1 ?
      SELECTION_DIRECTION.BACKWARDS :
      SELECTION_DIRECTION.FORWARDS;
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-selectallchildren
  selectAllChildren(node) {
    if (node.nodeType === NODE_TYPE.DOCUMENT_TYPE_NODE) {
      throw DOMException.create(this._globalObject, [
        "DocumentType Node can't be used as boundary point.",
        "InvalidNodeTypeError"
      ]);
    ***REMOVED***

    const document = implForWrapper(this._globalObject._document);
    if (document !== nodeRoot(node)) {
      return;
    ***REMOVED***

    const length = domSymbolTree.childrenCount(node);

    const newRange = Range.createImpl(this._globalObject, [], {
      start: { node, offset: 0 ***REMOVED***,
      end: { node, offset: 0 ***REMOVED***
    ***REMOVED***);

    setBoundaryPointStart(newRange, node, 0);
    setBoundaryPointEnd(newRange, node, length);

    this._associateRange(newRange);
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-deletefromdocument
  deleteFromDocument() {
    if (!this._isEmpty()) {
      this._range.deleteContents();
    ***REMOVED***
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-containsnode
  containsNode(node, allowPartialContainment) {
    if (this._isEmpty() || nodeRoot(node) !== implForWrapper(this._globalObject._document)) {
      return false;
    ***REMOVED***

    const { _start, _end ***REMOVED*** = this._range;

    const startIsBeforeNode = compareBoundaryPointsPosition(_start, { node, offset: 0 ***REMOVED***) === -1;
    const endIsAfterNode = compareBoundaryPointsPosition(_end, { node, offset: nodeLength(node) ***REMOVED***) === 1;

    return allowPartialContainment ?
      startIsBeforeNode || endIsAfterNode :
      startIsBeforeNode && endIsAfterNode;
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dom-selection-stringifier
  toString() {
    return this._range ? this._range.toString() : "";
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dfn-empty
  _isEmpty() {
    return this._range === null;
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dfn-anchor
  get _anchor() {
    if (!this._range) {
      return null;
    ***REMOVED***

    return this._direction === SELECTION_DIRECTION.FORWARDS ?
      this._range._start :
      this._range._end;
  ***REMOVED***

  // https://w3c.github.io/selection-api/#dfn-focus
  get _focus() {
    if (!this._range) {
      return null;
    ***REMOVED***

    return this._direction === SELECTION_DIRECTION.FORWARDS ?
      this._range._end :
      this._range._start;
  ***REMOVED***

  _associateRange(newRange) {
    this._range = newRange;
    this._direction = newRange === null ? SELECTION_DIRECTION.DIRECTIONLESS : SELECTION_DIRECTION.FORWARDS;

    // TODO: Emit "selectionchange" event. At this time, there is currently no test in WPT covering this.
    // https://w3c.github.io/selection-api/#selectionchange-event
  ***REMOVED***
***REMOVED***

module.exports = {
  implementation: SelectionImpl
***REMOVED***;
