// istanbul-lib-instrument/src/read-coverage.js
const { parseSync, traverse ***REMOVED*** = require('@babel/core');
const { defaults ***REMOVED*** = require('@istanbuljs/schema');
const { MAGIC_KEY, MAGIC_VALUE ***REMOVED*** = require('./constants');

function getAst(code) {
    if (typeof code === 'object' && typeof code.type === 'string') {
        // Assume code is already a babel ast.
        return code;
    ***REMOVED***

    if (typeof code !== 'string') {
        throw new Error('Code must be a string');
    ***REMOVED***

    // Parse as leniently as possible
    return parseSync(code, {
        babelrc: false,
        configFile: false,
        parserOpts: {
            allowAwaitOutsideFunction: true,
            allowImportExportEverywhere: true,
            allowReturnOutsideFunction: true,
            allowSuperOutsideMethod: true,
            sourceType: 'script',
            plugins: defaults.instrumenter.parserPlugins
        ***REMOVED***
    ***REMOVED***);
***REMOVED***

module.exports = function readInitialCoverage(code) {
    const ast = getAst(code);

    let covScope;
    traverse(ast, {
        ObjectProperty(path) {
            const { node ***REMOVED*** = path;
            if (
                !node.computed &&
                path.get('key').isIdentifier() &&
                node.key.name === MAGIC_KEY
            ) {
                const magicValue = path.get('value').evaluate();
                if (!magicValue.confident || magicValue.value !== MAGIC_VALUE) {
                    return;
                ***REMOVED***
                covScope =
                    path.scope.getFunctionParent() ||
                    path.scope.getProgramParent();
                path.stop();
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***);

    if (!covScope) {
        return null;
    ***REMOVED***

    const result = {***REMOVED***;

    for (const key of ['path', 'hash', 'gcv', 'coverageData']) {
        const binding = covScope.getOwnBinding(key);
        if (!binding) {
            return null;
        ***REMOVED***
        const valuePath = binding.path.get('init');
        const value = valuePath.evaluate();
        if (!value.confident) {
            return null;
        ***REMOVED***
        result[key] = value.value;
    ***REMOVED***

    delete result.coverageData[MAGIC_KEY];
    delete result.coverageData.hash;

    return result;
***REMOVED***;
