// jest-watch-typeahead/build/test_name_plugin/plugin.js
import { Prompt ***REMOVED*** from 'jest-watcher';
import TestNamePatternPrompt from "./prompt.js";
export default class TestNamePlugin {
  constructor({
    stdin,
    stdout,
    config = {***REMOVED***
  ***REMOVED***) {
    this._stdin = stdin;
    this._stdout = stdout;
    this._prompt = new Prompt();
    this._testResults = [];
    this._usageInfo = {
      key: config.key || 't',
      prompt: config.prompt || 'filter by a test name regex pattern'
    ***REMOVED***;
  ***REMOVED***

  apply(jestHooks) {
    jestHooks.onTestRunComplete(({
      testResults
    ***REMOVED***) => {
      this._testResults = testResults;
    ***REMOVED***);
  ***REMOVED***

  onKey(key) {
    this._prompt.put(key);
  ***REMOVED***

  run(globalConfig, updateConfigAndRun) {
    const p = new TestNamePatternPrompt(this._stdout, this._prompt);
    p.updateCachedTestResults(this._testResults);
    return new Promise((res, rej) => {
      p.run(testNamePattern => {
        updateConfigAndRun({
          mode: 'watch',
          testNamePattern
        ***REMOVED***);
        res();
      ***REMOVED***, rej);
    ***REMOVED***);
  ***REMOVED***

  getUsageInfo() {
    return this._usageInfo;
  ***REMOVED***

***REMOVED***
