// @babel/helper-plugin-utils/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.declare = declare;
exports.declarePreset = void 0;
const apiPolyfills = {
  assertVersion: api => range => {
    throwVersionError(range, api.version);
  ***REMOVED***
***REMOVED***;
{
  Object.assign(apiPolyfills, {
    targets: () => () => {
      return {***REMOVED***;
    ***REMOVED***,
    assumption: () => () => {
      return undefined;
    ***REMOVED***,
    addExternalDependency: () => () => {***REMOVED***
  ***REMOVED***);
***REMOVED***
function declare(builder) {
  return (api, options, dirname) => {
    let clonedApi;
    for (const name of Object.keys(apiPolyfills)) {
      if (api[name]) continue;
      clonedApi != null ? clonedApi : clonedApi = copyApiObject(api);
      clonedApi[name] = apiPolyfills[name](clonedApi);
    ***REMOVED***
    return builder(clonedApi != null ? clonedApi : api, options || {***REMOVED***, dirname);
  ***REMOVED***;
***REMOVED***
const declarePreset = exports.declarePreset = declare;
function copyApiObject(api) {
  let proto = null;
  if (typeof api.version === "string" && /^7\./.test(api.version)) {
    proto = Object.getPrototypeOf(api);
    if (proto && (!hasOwnProperty.call(proto, "version") || !hasOwnProperty.call(proto, "transform") || !hasOwnProperty.call(proto, "template") || !hasOwnProperty.call(proto, "types"))) {
      proto = null;
    ***REMOVED***
  ***REMOVED***
  return Object.assign({***REMOVED***, proto, api);
***REMOVED***
function throwVersionError(range, version) {
  if (typeof range === "number") {
    if (!Number.isInteger(range)) {
      throw new Error("Expected string or integer value.");
    ***REMOVED***
    range = `^${range***REMOVED***.0.0-0`;
  ***REMOVED***
  if (typeof range !== "string") {
    throw new Error("Expected string or integer value.");
  ***REMOVED***
  const limit = Error.stackTraceLimit;
  if (typeof limit === "number" && limit < 25) {
    Error.stackTraceLimit = 25;
  ***REMOVED***
  let err;
  if (version.slice(0, 2) === "7.") {
    err = new Error(`Requires Babel "^7.0.0-beta.41", but was loaded with "${version***REMOVED***". ` + `You'll need to update your @babel/core version.`);
  ***REMOVED*** else {
    err = new Error(`Requires Babel "${range***REMOVED***", but was loaded with "${version***REMOVED***". ` + `If you are sure you have a compatible version of @babel/core, ` + `it is likely that something in your build process is loading the ` + `wrong version. Inspect the stack trace of this error to look for ` + `the first entry that doesn't mention "@babel/core" or "babel-core" ` + `to see what is calling Babel.`);
  ***REMOVED***
  if (typeof limit === "number") {
    Error.stackTraceLimit = limit;
  ***REMOVED***
  throw Object.assign(err, {
    code: "BABEL_VERSION_UNSUPPORTED",
    version,
    range
  ***REMOVED***);
***REMOVED***
