// yargs/lib/platform-shims/browser.mjs
'use strict';

import cliui from 'https://unpkg.com/cliui@7.0.1/index.mjs'; // eslint-disable-line
import Parser from 'https://unpkg.com/yargs-parser@19.0.0/browser.js'; // eslint-disable-line
import {getProcessArgvBin***REMOVED*** from '../../build/lib/utils/process-argv.js';
import {YError***REMOVED*** from '../../build/lib/yerror.js';

const REQUIRE_ERROR = 'require is not supported in browser';
const REQUIRE_DIRECTORY_ERROR =
  'loading a directory of commands is not supported in browser';

export default {
  assert: {
    notStrictEqual: (a, b) => {
      // noop.
    ***REMOVED***,
    strictEqual: (a, b) => {
      // noop.
    ***REMOVED***,
  ***REMOVED***,
  cliui,
  findUp: () => undefined,
  getEnv: key => {
    // There is no environment in browser:
    return undefined;
  ***REMOVED***,
  inspect: console.log,
  getCallerFile: () => {
    throw new YError(REQUIRE_DIRECTORY_ERROR);
  ***REMOVED***,
  getProcessArgvBin,
  mainFilename: 'yargs',
  Parser,
  path: {
    basename: str => str,
    dirname: str => str,
    extname: str => str,
    relative: str => str,
  ***REMOVED***,
  process: {
    argv: () => [],
    cwd: () => '',
    execPath: () => '',
    // exit is noop browser:
    exit: () => {***REMOVED***,
    nextTick: cb => {
      window.setTimeout(cb, 1);
    ***REMOVED***,
    stdColumns: 80,
  ***REMOVED***,
  readFileSync: () => {
    return '';
  ***REMOVED***,
  require: () => {
    throw new YError(REQUIRE_ERROR);
  ***REMOVED***,
  requireDirectory: () => {
    throw new YError(REQUIRE_DIRECTORY_ERROR);
  ***REMOVED***,
  stringWidth: str => {
    return [...str].length;
  ***REMOVED***,
  // TODO: replace this with y18n once it's ported to ESM:
  y18n: {
    __: (...str) => {
      if (str.length === 0) return '';
      const args = str.slice(1);
      return sprintf(str[0], ...args);
    ***REMOVED***,
    __n: (str1, str2, count, ...args) => {
      if (count === 1) {
        return sprintf(str1, ...args);
      ***REMOVED*** else {
        return sprintf(str2, ...args);
      ***REMOVED***
    ***REMOVED***,
    getLocale: () => {
      return 'en_US';
    ***REMOVED***,
    setLocale: () => {***REMOVED***,
    updateLocale: () => {***REMOVED***,
  ***REMOVED***,
***REMOVED***;

function sprintf(_str, ...args) {
  let str = '';
  const split = _str.split('%s');
  split.forEach((token, i) => {
    str += `${token***REMOVED***${split[i + 1] !== undefined && args[i] ? args[i] : ''***REMOVED***`;
  ***REMOVED***);
  return str;
***REMOVED***
