// yargs/lib/platform-shims/esm.mjs
'use strict'

import { notStrictEqual, strictEqual ***REMOVED*** from 'assert'
import cliui from 'cliui'
import escalade from 'escalade/sync'
import { format, inspect ***REMOVED*** from 'util'
import { readFileSync ***REMOVED*** from 'fs'
import { fileURLToPath ***REMOVED*** from 'url';
import Parser from 'yargs-parser'
import { basename, dirname, extname, relative, resolve ***REMOVED*** from 'path'
import { getProcessArgvBin ***REMOVED*** from '../../build/lib/utils/process-argv.js'
import { YError ***REMOVED*** from '../../build/lib/yerror.js'
import y18n from 'y18n'

const REQUIRE_ERROR = 'require is not supported by ESM'
const REQUIRE_DIRECTORY_ERROR = 'loading a directory of commands is not supported yet for ESM'

const mainFilename = fileURLToPath(import.meta.url).split('node_modules')[0]
const __dirname = fileURLToPath(import.meta.url)

export default {
  assert: {
    notStrictEqual,
    strictEqual
  ***REMOVED***,
  cliui,
  findUp: escalade,
  getEnv: (key) => {
    return process.env[key]
  ***REMOVED***,
  inspect,
  getCallerFile: () => {
    throw new YError(REQUIRE_DIRECTORY_ERROR)
  ***REMOVED***,
  getProcessArgvBin,
  mainFilename: mainFilename || process.cwd(),
  Parser,
  path: {
    basename,
    dirname,
    extname,
    relative,
    resolve
  ***REMOVED***,
  process: {
    argv: () => process.argv,
    cwd: process.cwd,
    execPath: () => process.execPath,
    exit: process.exit,
    nextTick: process.nextTick,
    stdColumns: typeof process.stdout.columns !== 'undefined' ? process.stdout.columns : null
  ***REMOVED***,
  readFileSync,
  require: () => {
    throw new YError(REQUIRE_ERROR)
  ***REMOVED***,
  requireDirectory: () => {
    throw new YError(REQUIRE_DIRECTORY_ERROR)
  ***REMOVED***,
  stringWidth: (str) => {
    return [...str].length
  ***REMOVED***,
  y18n: y18n({
    directory: resolve(__dirname, '../../../locales'),
    updateFiles: false
  ***REMOVED***)
***REMOVED***
