// webpack-dev-server/client/overlay/state-machine.js
import createMachine from "./fsm.js";

/**
 * @typedef {Object***REMOVED*** ShowOverlayData
 * @property {'warning' | 'error'***REMOVED*** level
 * @property {Array<string  | { moduleIdentifier?: string, moduleName?: string, loc?: string, message?: string ***REMOVED***>***REMOVED*** messages
 * @property {'build' | 'runtime'***REMOVED*** messageSource
 */

/**
 * @typedef {Object***REMOVED*** CreateOverlayMachineOptions
 * @property {(data: ShowOverlayData) => void***REMOVED*** showOverlay
 * @property {() => void***REMOVED*** hideOverlay
 */

/**
 * @param {CreateOverlayMachineOptions***REMOVED*** options
 */
var createOverlayMachine = function createOverlayMachine(options) {
  var hideOverlay = options.hideOverlay,
    showOverlay = options.showOverlay;
  var overlayMachine = createMachine({
    initial: "hidden",
    context: {
      level: "error",
      messages: [],
      messageSource: "build"
    ***REMOVED***,
    states: {
      hidden: {
        on: {
          BUILD_ERROR: {
            target: "displayBuildError",
            actions: ["setMessages", "showOverlay"]
          ***REMOVED***,
          RUNTIME_ERROR: {
            target: "displayRuntimeError",
            actions: ["setMessages", "showOverlay"]
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***,
      displayBuildError: {
        on: {
          DISMISS: {
            target: "hidden",
            actions: ["dismissMessages", "hideOverlay"]
          ***REMOVED***,
          BUILD_ERROR: {
            target: "displayBuildError",
            actions: ["appendMessages", "showOverlay"]
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***,
      displayRuntimeError: {
        on: {
          DISMISS: {
            target: "hidden",
            actions: ["dismissMessages", "hideOverlay"]
          ***REMOVED***,
          RUNTIME_ERROR: {
            target: "displayRuntimeError",
            actions: ["appendMessages", "showOverlay"]
          ***REMOVED***,
          BUILD_ERROR: {
            target: "displayBuildError",
            actions: ["setMessages", "showOverlay"]
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***, {
    actions: {
      dismissMessages: function dismissMessages() {
        return {
          messages: [],
          level: "error",
          messageSource: "build"
        ***REMOVED***;
      ***REMOVED***,
      appendMessages: function appendMessages(context, event) {
        return {
          messages: context.messages.concat(event.messages),
          level: event.level || context.level,
          messageSource: event.type === "RUNTIME_ERROR" ? "runtime" : "build"
        ***REMOVED***;
      ***REMOVED***,
      setMessages: function setMessages(context, event) {
        return {
          messages: event.messages,
          level: event.level || context.level,
          messageSource: event.type === "RUNTIME_ERROR" ? "runtime" : "build"
        ***REMOVED***;
      ***REMOVED***,
      hideOverlay: hideOverlay,
      showOverlay: showOverlay
    ***REMOVED***
  ***REMOVED***);
  return overlayMachine;
***REMOVED***;
export default createOverlayMachine;
