// @svgr/babel-plugin-svg-dynamic-title/lib/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;
const elements = ['svg', 'Svg'];

const plugin = ({
  types: t
***REMOVED***) => ({
  visitor: {
    JSXElement(path) {
      if (!elements.some(element => path.get('openingElement.name').isJSXIdentifier({
        name: element
      ***REMOVED***))) {
        return;
      ***REMOVED***

      function createTitle(children = [], attributes = []) {
        return t.jsxElement(t.jsxOpeningElement(t.jsxIdentifier('title'), attributes), t.jsxClosingElement(t.jsxIdentifier('title')), children);
      ***REMOVED***

      function createTitleIdAttribute() {
        return t.jsxAttribute(t.jsxIdentifier('id'), t.jsxExpressionContainer(t.identifier('titleId')));
      ***REMOVED***

      function enhanceAttributes(attributes) {
        const existingId = attributes.find(attribute => attribute.name.name === 'id');

        if (!existingId) {
          return [...attributes, createTitleIdAttribute()];
        ***REMOVED***

        existingId.value = t.jsxExpressionContainer(t.logicalExpression('||', t.identifier('titleId'), existingId.value));
        return attributes;
      ***REMOVED***

      function getTitleElement(existingTitle) {
        const titleExpression = t.identifier('title');

        if (existingTitle) {
          existingTitle.openingElement.attributes = enhanceAttributes(existingTitle.openingElement.attributes);
        ***REMOVED***

        let titleElement = t.conditionalExpression(titleExpression, createTitle([t.jsxExpressionContainer(titleExpression)], existingTitle ? existingTitle.openingElement.attributes : [t.jsxAttribute(t.jsxIdentifier('id'), t.jsxExpressionContainer(t.identifier('titleId')))]), t.nullLiteral());

        if (existingTitle && existingTitle.children && existingTitle.children.length) {
          // if title already exists
          // render as follows
          const fallbackTitleElement = existingTitle; // {title === undefined ? fallbackTitleElement : titleElement***REMOVED***

          const conditionalExpressionForTitle = t.conditionalExpression(t.binaryExpression('===', titleExpression, t.identifier('undefined')), fallbackTitleElement, titleElement);
          titleElement = t.jsxExpressionContainer(conditionalExpressionForTitle);
        ***REMOVED*** else {
          titleElement = t.jsxExpressionContainer(titleElement);
        ***REMOVED***

        return titleElement;
      ***REMOVED*** // store the title element


      let titleElement;
      const hasTitle = path.get('children').some(childPath => {
        if (!childPath.isJSXElement()) return false;
        if (childPath.node === titleElement) return false;
        if (childPath.node.openingElement.name.name !== 'title') return false;
        titleElement = getTitleElement(childPath.node);
        childPath.replaceWith(titleElement);
        return true;
      ***REMOVED***); // create a title element if not already create

      titleElement = titleElement || getTitleElement();

      if (!hasTitle) {
        // path.unshiftContainer is not working well :(
        // path.unshiftContainer('children', titleElement)
        path.node.children.unshift(titleElement);
        path.replaceWith(path.node);
      ***REMOVED***
    ***REMOVED***

  ***REMOVED***
***REMOVED***);

var _default = plugin;
exports.default = _default;
