// ajv/lib/vocabularies/jtd/values.ts
import type {CodeKeywordDefinition, SchemaObject***REMOVED*** from "../../types"
import type {KeywordCxt***REMOVED*** from "../../compile/validate"
import {alwaysValidSchema, Type***REMOVED*** from "../../compile/util"
import {not, or, Name***REMOVED*** from "../../compile/codegen"
import {checkMetadata***REMOVED*** from "./metadata"
import {checkNullableObject***REMOVED*** from "./nullable"
import {typeError, _JTDTypeError***REMOVED*** from "./error"

export type JTDValuesError = _JTDTypeError<"values", "object", SchemaObject>

const def: CodeKeywordDefinition = {
  keyword: "values",
  schemaType: "object",
  error: typeError("object"),
  code(cxt: KeywordCxt) {
    checkMetadata(cxt)
    const {gen, data, schema, it***REMOVED*** = cxt
    const [valid, cond] = checkNullableObject(cxt, data)
    if (alwaysValidSchema(it, schema)) {
      gen.if(not(or(cond, valid)), () => cxt.error())
    ***REMOVED*** else {
      gen.if(cond)
      gen.assign(valid, validateMap())
      gen.elseIf(not(valid))
      cxt.error()
      gen.endIf()
    ***REMOVED***
    cxt.ok(valid)

    function validateMap(): Name | boolean {
      const _valid = gen.name("valid")
      if (it.allErrors) {
        const validMap = gen.let("valid", true)
        validateValues(() => gen.assign(validMap, false))
        return validMap
      ***REMOVED***
      gen.var(_valid, true)
      validateValues(() => gen.break())
      return _valid

      function validateValues(notValid: () => void): void {
        gen.forIn("key", data, (key) => {
          cxt.subschema(
            {
              keyword: "values",
              dataProp: key,
              dataPropType: Type.Str,
            ***REMOVED***,
            _valid
          )
          gen.if(not(_valid), notValid)
        ***REMOVED***)
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***,
***REMOVED***

export default def
