// postcss-calc/src/index.js
'use strict';
const transform = require('./lib/transform.js');

/**
 * @typedef {{precision?: number | false,
 *          preserve?: boolean,
 *          warnWhenCannotResolve?: boolean,
 *          mediaQueries?: boolean,
 *          selectors?: boolean***REMOVED******REMOVED*** PostCssCalcOptions
 */
/**
 * @type {import('postcss').PluginCreator<PostCssCalcOptions>***REMOVED***
 * @param {PostCssCalcOptions***REMOVED*** opts
 * @return {import('postcss').Plugin***REMOVED***
 */
function pluginCreator(opts) {
  const options = Object.assign(
    {
      precision: 5,
      preserve: false,
      warnWhenCannotResolve: false,
      mediaQueries: false,
      selectors: false,
    ***REMOVED***,
    opts
  );

  return {
    postcssPlugin: 'postcss-calc',
    OnceExit(css, { result ***REMOVED***) {
      css.walk((node) => {
        const { type ***REMOVED*** = node;
        if (type === 'decl') {
          transform(node, 'value', options, result);
        ***REMOVED***

        if (type === 'atrule' && options.mediaQueries) {
          transform(node, 'params', options, result);
        ***REMOVED***

        if (type === 'rule' && options.selectors) {
          transform(node, 'selector', options, result);
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***,
  ***REMOVED***;
***REMOVED***

pluginCreator.postcss = true;

module.exports = pluginCreator;
