// tailwindcss/lib/corePlugins.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
***REMOVED***);
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    ***REMOVED***);
***REMOVED***
_export(exports, {
    variantPlugins: function() {
        return variantPlugins;
    ***REMOVED***,
    corePlugins: function() {
        return corePlugins;
    ***REMOVED***
***REMOVED***);
const _fs = /*#__PURE__*/ _interop_require_default(require("fs"));
const _path = /*#__PURE__*/ _interop_require_wildcard(require("path"));
const _postcss = /*#__PURE__*/ _interop_require_default(require("postcss"));
const _createUtilityPlugin = /*#__PURE__*/ _interop_require_default(require("./util/createUtilityPlugin"));
const _buildMediaQuery = /*#__PURE__*/ _interop_require_default(require("./util/buildMediaQuery"));
const _escapeClassName = /*#__PURE__*/ _interop_require_default(require("./util/escapeClassName"));
const _parseAnimationValue = /*#__PURE__*/ _interop_require_default(require("./util/parseAnimationValue"));
const _flattenColorPalette = /*#__PURE__*/ _interop_require_default(require("./util/flattenColorPalette"));
const _withAlphaVariable = /*#__PURE__*/ _interop_require_wildcard(require("./util/withAlphaVariable"));
const _toColorValue = /*#__PURE__*/ _interop_require_default(require("./util/toColorValue"));
const _isPlainObject = /*#__PURE__*/ _interop_require_default(require("./util/isPlainObject"));
const _transformThemeValue = /*#__PURE__*/ _interop_require_default(require("./util/transformThemeValue"));
const _packagejson = require("../package.json");
const _log = /*#__PURE__*/ _interop_require_default(require("./util/log"));
const _normalizeScreens = require("./util/normalizeScreens");
const _parseBoxShadowValue = require("./util/parseBoxShadowValue");
const _removeAlphaVariables = require("./util/removeAlphaVariables");
const _featureFlags = require("./featureFlags");
const _dataTypes = require("./util/dataTypes");
const _setupContextUtils = require("./lib/setupContextUtils");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    ***REMOVED***;
***REMOVED***
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    ***REMOVED***)(nodeInterop);
***REMOVED***
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    ***REMOVED***
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        ***REMOVED***;
    ***REMOVED***
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    ***REMOVED***
    var newObj = {***REMOVED***;
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            ***REMOVED*** else {
                newObj[key] = obj[key];
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    ***REMOVED***
    return newObj;
***REMOVED***
let variantPlugins = {
    childVariant: ({ addVariant  ***REMOVED***)=>{
        addVariant("*", "& > *");
    ***REMOVED***,
    pseudoElementVariants: ({ addVariant  ***REMOVED***)=>{
        addVariant("first-letter", "&::first-letter");
        addVariant("first-line", "&::first-line");
        addVariant("marker", [
            ({ container  ***REMOVED***)=>{
                (0, _removeAlphaVariables.removeAlphaVariables)(container, [
                    "--tw-text-opacity"
                ]);
                return "& *::marker";
            ***REMOVED***,
            ({ container  ***REMOVED***)=>{
                (0, _removeAlphaVariables.removeAlphaVariables)(container, [
                    "--tw-text-opacity"
                ]);
                return "&::marker";
            ***REMOVED***
        ]);
        addVariant("selection", [
            "& *::selection",
            "&::selection"
        ]);
        addVariant("file", "&::file-selector-button");
        addVariant("placeholder", "&::placeholder");
        addVariant("backdrop", "&::backdrop");
        addVariant("before", ({ container  ***REMOVED***)=>{
            container.walkRules((rule)=>{
                let foundContent = false;
                rule.walkDecls("content", ()=>{
                    foundContent = true;
                ***REMOVED***);
                if (!foundContent) {
                    rule.prepend(_postcss.default.decl({
                        prop: "content",
                        value: "var(--tw-content)"
                    ***REMOVED***));
                ***REMOVED***
            ***REMOVED***);
            return "&::before";
        ***REMOVED***);
        addVariant("after", ({ container  ***REMOVED***)=>{
            container.walkRules((rule)=>{
                let foundContent = false;
                rule.walkDecls("content", ()=>{
                    foundContent = true;
                ***REMOVED***);
                if (!foundContent) {
                    rule.prepend(_postcss.default.decl({
                        prop: "content",
                        value: "var(--tw-content)"
                    ***REMOVED***));
                ***REMOVED***
            ***REMOVED***);
            return "&::after";
        ***REMOVED***);
    ***REMOVED***,
    pseudoClassVariants: ({ addVariant , matchVariant , config , prefix  ***REMOVED***)=>{
        let pseudoVariants = [
            // Positional
            [
                "first",
                "&:first-child"
            ],
            [
                "last",
                "&:last-child"
            ],
            [
                "only",
                "&:only-child"
            ],
            [
                "odd",
                "&:nth-child(odd)"
            ],
            [
                "even",
                "&:nth-child(even)"
            ],
            "first-of-type",
            "last-of-type",
            "only-of-type",
            // State
            [
                "visited",
                ({ container  ***REMOVED***)=>{
                    (0, _removeAlphaVariables.removeAlphaVariables)(container, [
                        "--tw-text-opacity",
                        "--tw-border-opacity",
                        "--tw-bg-opacity"
                    ]);
                    return "&:visited";
                ***REMOVED***
            ],
            "target",
            [
                "open",
                "&[open]"
            ],
            // Forms
            "default",
            "checked",
            "indeterminate",
            "placeholder-shown",
            "autofill",
            "optional",
            "required",
            "valid",
            "invalid",
            "in-range",
            "out-of-range",
            "read-only",
            // Content
            "empty",
            // Interactive
            "focus-within",
            [
                "hover",
                !(0, _featureFlags.flagEnabled)(config(), "hoverOnlyWhenSupported") ? "&:hover" : "@media (hover: hover) and (pointer: fine) { &:hover ***REMOVED***"
            ],
            "focus",
            "focus-visible",
            "active",
            "enabled",
            "disabled"
        ].map((variant)=>Array.isArray(variant) ? variant : [
                variant,
                `&:${variant***REMOVED***`
            ]);
        for (let [variantName, state] of pseudoVariants){
            addVariant(variantName, (ctx)=>{
                let result = typeof state === "function" ? state(ctx) : state;
                return result;
            ***REMOVED***);
        ***REMOVED***
        let variants = {
            group: (_, { modifier  ***REMOVED***)=>modifier ? [
                    `:merge(${prefix(".group")***REMOVED***\\/${(0, _escapeClassName.default)(modifier)***REMOVED***)`,
                    " &"
                ] : [
                    `:merge(${prefix(".group")***REMOVED***)`,
                    " &"
                ],
            peer: (_, { modifier  ***REMOVED***)=>modifier ? [
                    `:merge(${prefix(".peer")***REMOVED***\\/${(0, _escapeClassName.default)(modifier)***REMOVED***)`,
                    " ~ &"
                ] : [
                    `:merge(${prefix(".peer")***REMOVED***)`,
                    " ~ &"
                ]
        ***REMOVED***;
        for (let [name, fn] of Object.entries(variants)){
            matchVariant(name, (value = "", extra)=>{
                let result = (0, _dataTypes.normalize)(typeof value === "function" ? value(extra) : value);
                if (!result.includes("&")) result = "&" + result;
                let [a, b] = fn("", extra);
                let start = null;
                let end = null;
                let quotes = 0;
                for(let i = 0; i < result.length; ++i){
                    let c = result[i];
                    if (c === "&") {
                        start = i;
                    ***REMOVED*** else if (c === "'" || c === '"') {
                        quotes += 1;
                    ***REMOVED*** else if (start !== null && c === " " && !quotes) {
                        end = i;
                    ***REMOVED***
                ***REMOVED***
                if (start !== null && end === null) {
                    end = result.length;
                ***REMOVED***
                // Basically this but can handle quotes:
                // result.replace(/&(\S+)?/g, (_, pseudo = '') => a + pseudo + b)
                return result.slice(0, start) + a + result.slice(start + 1, end) + b + result.slice(end);
            ***REMOVED***, {
                values: Object.fromEntries(pseudoVariants),
                [_setupContextUtils.INTERNAL_FEATURES]: {
                    respectPrefix: false
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***,
    directionVariants: ({ addVariant  ***REMOVED***)=>{
        addVariant("ltr", '&:where([dir="ltr"], [dir="ltr"] *)');
        addVariant("rtl", '&:where([dir="rtl"], [dir="rtl"] *)');
    ***REMOVED***,
    reducedMotionVariants: ({ addVariant  ***REMOVED***)=>{
        addVariant("motion-safe", "@media (prefers-reduced-motion: no-preference)");
        addVariant("motion-reduce", "@media (prefers-reduced-motion: reduce)");
    ***REMOVED***,
    darkVariants: ({ config , addVariant  ***REMOVED***)=>{
        let [mode, selector = ".dark"] = [].concat(config("darkMode", "media"));
        if (mode === false) {
            mode = "media";
            _log.default.warn("darkmode-false", [
                "The `darkMode` option in your Tailwind CSS configuration is set to `false`, which now behaves the same as `media`.",
                "Change `darkMode` to `media` or remove it entirely.",
                "https://tailwindcss.com/docs/upgrade-guide#remove-dark-mode-configuration"
            ]);
        ***REMOVED***
        if (mode === "variant") {
            let formats;
            if (Array.isArray(selector)) {
                formats = selector;
            ***REMOVED*** else if (typeof selector === "function") {
                formats = selector;
            ***REMOVED*** else if (typeof selector === "string") {
                formats = [
                    selector
                ];
            ***REMOVED***
            // TODO: We could also add these warnings if the user passes a function that returns string | string[]
            // But this is an advanced enough use case that it's probably not necessary
            if (Array.isArray(formats)) {
                for (let format of formats){
                    if (format === ".dark") {
                        mode = false;
                        _log.default.warn("darkmode-variant-without-selector", [
                            "When using `variant` for `darkMode`, you must provide a selector.",
                            'Example: `darkMode: ["variant", ".your-selector &"]`'
                        ]);
                    ***REMOVED*** else if (!format.includes("&")) {
                        mode = false;
                        _log.default.warn("darkmode-variant-without-ampersand", [
                            "When using `variant` for `darkMode`, your selector must contain `&`.",
                            'Example `darkMode: ["variant", ".your-selector &"]`'
                        ]);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
            selector = formats;
        ***REMOVED***
        if (mode === "selector") {
            // New preferred behavior
            addVariant("dark", `&:where(${selector***REMOVED***, ${selector***REMOVED*** *)`);
        ***REMOVED*** else if (mode === "media") {
            addVariant("dark", "@media (prefers-color-scheme: dark)");
        ***REMOVED*** else if (mode === "variant") {
            addVariant("dark", selector);
        ***REMOVED*** else if (mode === "class") {
            // Old behavior
            addVariant("dark", `&:is(${selector***REMOVED*** *)`);
        ***REMOVED***
    ***REMOVED***,
    printVariant: ({ addVariant  ***REMOVED***)=>{
        addVariant("print", "@media print");
    ***REMOVED***,
    screenVariants: ({ theme , addVariant , matchVariant  ***REMOVED***)=>{
        var _theme;
        let rawScreens = (_theme = theme("screens")) !== null && _theme !== void 0 ? _theme : {***REMOVED***;
        let areSimpleScreens = Object.values(rawScreens).every((v)=>typeof v === "string");
        let screens = (0, _normalizeScreens.normalizeScreens)(theme("screens"));
        /** @type {Set<string>***REMOVED*** */ let unitCache = new Set([]);
        /** @param {string***REMOVED*** value */ function units(value) {
            var _value_match;
            var _value_match_;
            return (_value_match_ = (_value_match = value.match(/(\D+)$/)) === null || _value_match === void 0 ? void 0 : _value_match[1]) !== null && _value_match_ !== void 0 ? _value_match_ : "(none)";
        ***REMOVED***
        /** @param {string***REMOVED*** value */ function recordUnits(value) {
            if (value !== undefined) {
                unitCache.add(units(value));
            ***REMOVED***
        ***REMOVED***
        /** @param {string***REMOVED*** value */ function canUseUnits(value) {
            recordUnits(value);
            // If the cache was empty it'll become 1 because we've just added the current unit
            // If the cache was not empty and the units are the same the size doesn't change
            // Otherwise, if the units are different from what is already known the size will always be > 1
            return unitCache.size === 1;
        ***REMOVED***
        for (const screen of screens){
            for (const value of screen.values){
                recordUnits(value.min);
                recordUnits(value.max);
            ***REMOVED***
        ***REMOVED***
        let screensUseConsistentUnits = unitCache.size <= 1;
        /**
     * @typedef {import('./util/normalizeScreens').Screen***REMOVED*** Screen
     */ /**
     * @param {'min' | 'max'***REMOVED*** type
     * @returns {Record<string, Screen>***REMOVED***
     */ function buildScreenValues(type) {
            return Object.fromEntries(screens.filter((screen)=>(0, _normalizeScreens.isScreenSortable)(screen).result).map((screen)=>{
                let { min , max  ***REMOVED*** = screen.values[0];
                if (type === "min" && min !== undefined) {
                    return screen;
                ***REMOVED*** else if (type === "min" && max !== undefined) {
                    return {
                        ...screen,
                        not: !screen.not
                    ***REMOVED***;
                ***REMOVED*** else if (type === "max" && max !== undefined) {
                    return screen;
                ***REMOVED*** else if (type === "max" && min !== undefined) {
                    return {
                        ...screen,
                        not: !screen.not
                    ***REMOVED***;
                ***REMOVED***
            ***REMOVED***).map((screen)=>[
                    screen.name,
                    screen
                ]));
        ***REMOVED***
        /**
     * @param {'min' | 'max'***REMOVED*** type
     * @returns {(a: { value: string | Screen ***REMOVED***, z: { value: string | Screen ***REMOVED***) => number***REMOVED***
     */ function buildSort(type) {
            return (a, z)=>(0, _normalizeScreens.compareScreens)(type, a.value, z.value);
        ***REMOVED***
        let maxSort = buildSort("max");
        let minSort = buildSort("min");
        /** @param {'min'|'max'***REMOVED*** type */ function buildScreenVariant(type) {
            return (value)=>{
                if (!areSimpleScreens) {
                    _log.default.warn("complex-screen-config", [
                        "The `min-*` and `max-*` variants are not supported with a `screens` configuration containing objects."
                    ]);
                    return [];
                ***REMOVED*** else if (!screensUseConsistentUnits) {
                    _log.default.warn("mixed-screen-units", [
                        "The `min-*` and `max-*` variants are not supported with a `screens` configuration containing mixed units."
                    ]);
                    return [];
                ***REMOVED*** else if (typeof value === "string" && !canUseUnits(value)) {
                    _log.default.warn("minmax-have-mixed-units", [
                        "The `min-*` and `max-*` variants are not supported with a `screens` configuration containing mixed units."
                    ]);
                    return [];
                ***REMOVED***
                return [
                    `@media ${(0, _buildMediaQuery.default)((0, _normalizeScreens.toScreen)(value, type))***REMOVED***`
                ];
            ***REMOVED***;
        ***REMOVED***
        matchVariant("max", buildScreenVariant("max"), {
            sort: maxSort,
            values: areSimpleScreens ? buildScreenValues("max") : {***REMOVED***
        ***REMOVED***);
        // screens and min-* are sorted together when they can be
        let id = "min-screens";
        for (let screen of screens){
            addVariant(screen.name, `@media ${(0, _buildMediaQuery.default)(screen)***REMOVED***`, {
                id,
                sort: areSimpleScreens && screensUseConsistentUnits ? minSort : undefined,
                value: screen
            ***REMOVED***);
        ***REMOVED***
        matchVariant("min", buildScreenVariant("min"), {
            id,
            sort: minSort
        ***REMOVED***);
    ***REMOVED***,
    supportsVariants: ({ matchVariant , theme  ***REMOVED***)=>{
        var _theme;
        matchVariant("supports", (value = "")=>{
            let check = (0, _dataTypes.normalize)(value);
            let isRaw = /^\w*\s*\(/.test(check);
            // Chrome has a bug where `(condition1)or(condition2)` is not valid
            // But `(condition1) or (condition2)` is supported.
            check = isRaw ? check.replace(/\b(and|or|not)\b/g, " $1 ") : check;
            if (isRaw) {
                return `@supports ${check***REMOVED***`;
            ***REMOVED***
            if (!check.includes(":")) {
                check = `${check***REMOVED***: var(--tw)`;
            ***REMOVED***
            if (!(check.startsWith("(") && check.endsWith(")"))) {
                check = `(${check***REMOVED***)`;
            ***REMOVED***
            return `@supports ${check***REMOVED***`;
        ***REMOVED***, {
            values: (_theme = theme("supports")) !== null && _theme !== void 0 ? _theme : {***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    hasVariants: ({ matchVariant , prefix  ***REMOVED***)=>{
        matchVariant("has", (value)=>`&:has(${(0, _dataTypes.normalize)(value)***REMOVED***)`, {
            values: {***REMOVED***,
            [_setupContextUtils.INTERNAL_FEATURES]: {
                respectPrefix: false
            ***REMOVED***
        ***REMOVED***);
        matchVariant("group-has", (value, { modifier  ***REMOVED***)=>modifier ? `:merge(${prefix(".group")***REMOVED***\\/${modifier***REMOVED***):has(${(0, _dataTypes.normalize)(value)***REMOVED***) &` : `:merge(${prefix(".group")***REMOVED***):has(${(0, _dataTypes.normalize)(value)***REMOVED***) &`, {
            values: {***REMOVED***,
            [_setupContextUtils.INTERNAL_FEATURES]: {
                respectPrefix: false
            ***REMOVED***
        ***REMOVED***);
        matchVariant("peer-has", (value, { modifier  ***REMOVED***)=>modifier ? `:merge(${prefix(".peer")***REMOVED***\\/${modifier***REMOVED***):has(${(0, _dataTypes.normalize)(value)***REMOVED***) ~ &` : `:merge(${prefix(".peer")***REMOVED***):has(${(0, _dataTypes.normalize)(value)***REMOVED***) ~ &`, {
            values: {***REMOVED***,
            [_setupContextUtils.INTERNAL_FEATURES]: {
                respectPrefix: false
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    ariaVariants: ({ matchVariant , theme  ***REMOVED***)=>{
        var _theme;
        matchVariant("aria", (value)=>`&[aria-${(0, _dataTypes.normalizeAttributeSelectors)((0, _dataTypes.normalize)(value))***REMOVED***]`, {
            values: (_theme = theme("aria")) !== null && _theme !== void 0 ? _theme : {***REMOVED***
        ***REMOVED***);
        var _theme1;
        matchVariant("group-aria", (value, { modifier  ***REMOVED***)=>modifier ? `:merge(.group\\/${modifier***REMOVED***)[aria-${(0, _dataTypes.normalizeAttributeSelectors)((0, _dataTypes.normalize)(value))***REMOVED***] &` : `:merge(.group)[aria-${(0, _dataTypes.normalizeAttributeSelectors)((0, _dataTypes.normalize)(value))***REMOVED***] &`, {
            values: (_theme1 = theme("aria")) !== null && _theme1 !== void 0 ? _theme1 : {***REMOVED***
        ***REMOVED***);
        var _theme2;
        matchVariant("peer-aria", (value, { modifier  ***REMOVED***)=>modifier ? `:merge(.peer\\/${modifier***REMOVED***)[aria-${(0, _dataTypes.normalizeAttributeSelectors)((0, _dataTypes.normalize)(value))***REMOVED***] ~ &` : `:merge(.peer)[aria-${(0, _dataTypes.normalizeAttributeSelectors)((0, _dataTypes.normalize)(value))***REMOVED***] ~ &`, {
            values: (_theme2 = theme("aria")) !== null && _theme2 !== void 0 ? _theme2 : {***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    dataVariants: ({ matchVariant , theme  ***REMOVED***)=>{
        var _theme;
        matchVariant("data", (value)=>`&[data-${(0, _dataTypes.normalizeAttributeSelectors)((0, _dataTypes.normalize)(value))***REMOVED***]`, {
            values: (_theme = theme("data")) !== null && _theme !== void 0 ? _theme : {***REMOVED***
        ***REMOVED***);
        var _theme1;
        matchVariant("group-data", (value, { modifier  ***REMOVED***)=>modifier ? `:merge(.group\\/${modifier***REMOVED***)[data-${(0, _dataTypes.normalizeAttributeSelectors)((0, _dataTypes.normalize)(value))***REMOVED***] &` : `:merge(.group)[data-${(0, _dataTypes.normalizeAttributeSelectors)((0, _dataTypes.normalize)(value))***REMOVED***] &`, {
            values: (_theme1 = theme("data")) !== null && _theme1 !== void 0 ? _theme1 : {***REMOVED***
        ***REMOVED***);
        var _theme2;
        matchVariant("peer-data", (value, { modifier  ***REMOVED***)=>modifier ? `:merge(.peer\\/${modifier***REMOVED***)[data-${(0, _dataTypes.normalizeAttributeSelectors)((0, _dataTypes.normalize)(value))***REMOVED***] ~ &` : `:merge(.peer)[data-${(0, _dataTypes.normalizeAttributeSelectors)((0, _dataTypes.normalize)(value))***REMOVED***] ~ &`, {
            values: (_theme2 = theme("data")) !== null && _theme2 !== void 0 ? _theme2 : {***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    orientationVariants: ({ addVariant  ***REMOVED***)=>{
        addVariant("portrait", "@media (orientation: portrait)");
        addVariant("landscape", "@media (orientation: landscape)");
    ***REMOVED***,
    prefersContrastVariants: ({ addVariant  ***REMOVED***)=>{
        addVariant("contrast-more", "@media (prefers-contrast: more)");
        addVariant("contrast-less", "@media (prefers-contrast: less)");
    ***REMOVED***,
    forcedColorsVariants: ({ addVariant  ***REMOVED***)=>{
        addVariant("forced-colors", "@media (forced-colors: active)");
    ***REMOVED***
***REMOVED***;
let cssTransformValue = [
    "translate(var(--tw-translate-x), var(--tw-translate-y))",
    "rotate(var(--tw-rotate))",
    "skewX(var(--tw-skew-x))",
    "skewY(var(--tw-skew-y))",
    "scaleX(var(--tw-scale-x))",
    "scaleY(var(--tw-scale-y))"
].join(" ");
let cssFilterValue = [
    "var(--tw-blur)",
    "var(--tw-brightness)",
    "var(--tw-contrast)",
    "var(--tw-grayscale)",
    "var(--tw-hue-rotate)",
    "var(--tw-invert)",
    "var(--tw-saturate)",
    "var(--tw-sepia)",
    "var(--tw-drop-shadow)"
].join(" ");
let cssBackdropFilterValue = [
    "var(--tw-backdrop-blur)",
    "var(--tw-backdrop-brightness)",
    "var(--tw-backdrop-contrast)",
    "var(--tw-backdrop-grayscale)",
    "var(--tw-backdrop-hue-rotate)",
    "var(--tw-backdrop-invert)",
    "var(--tw-backdrop-opacity)",
    "var(--tw-backdrop-saturate)",
    "var(--tw-backdrop-sepia)"
].join(" ");
let corePlugins = {
    preflight: ({ addBase  ***REMOVED***)=>{
        let preflightStyles = _postcss.default.parse(_fs.default.readFileSync(_path.join(__dirname, "./css/preflight.css"), "utf8"));
        addBase([
            _postcss.default.comment({
                text: `! tailwindcss v${_packagejson.version***REMOVED*** | MIT License | https://tailwindcss.com`
            ***REMOVED***),
            ...preflightStyles.nodes
        ]);
    ***REMOVED***,
    container: (()=>{
        function extractMinWidths(breakpoints = []) {
            return breakpoints.flatMap((breakpoint)=>breakpoint.values.map((breakpoint)=>breakpoint.min)).filter((v)=>v !== undefined);
        ***REMOVED***
        function mapMinWidthsToPadding(minWidths, screens, paddings) {
            if (typeof paddings === "undefined") {
                return [];
            ***REMOVED***
            if (!(typeof paddings === "object" && paddings !== null)) {
                return [
                    {
                        screen: "DEFAULT",
                        minWidth: 0,
                        padding: paddings
                    ***REMOVED***
                ];
            ***REMOVED***
            let mapping = [];
            if (paddings.DEFAULT) {
                mapping.push({
                    screen: "DEFAULT",
                    minWidth: 0,
                    padding: paddings.DEFAULT
                ***REMOVED***);
            ***REMOVED***
            for (let minWidth of minWidths){
                for (let screen of screens){
                    for (let { min  ***REMOVED*** of screen.values){
                        if (min === minWidth) {
                            mapping.push({
                                minWidth,
                                padding: paddings[screen.name]
                            ***REMOVED***);
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
            return mapping;
        ***REMOVED***
        return function({ addComponents , theme  ***REMOVED***) {
            let screens = (0, _normalizeScreens.normalizeScreens)(theme("container.screens", theme("screens")));
            let minWidths = extractMinWidths(screens);
            let paddings = mapMinWidthsToPadding(minWidths, screens, theme("container.padding"));
            let generatePaddingFor = (minWidth)=>{
                let paddingConfig = paddings.find((padding)=>padding.minWidth === minWidth);
                if (!paddingConfig) {
                    return {***REMOVED***;
                ***REMOVED***
                return {
                    paddingRight: paddingConfig.padding,
                    paddingLeft: paddingConfig.padding
                ***REMOVED***;
            ***REMOVED***;
            let atRules = Array.from(new Set(minWidths.slice().sort((a, z)=>parseInt(a) - parseInt(z)))).map((minWidth)=>({
                    [`@media (min-width: ${minWidth***REMOVED***)`]: {
                        ".container": {
                            "max-width": minWidth,
                            ...generatePaddingFor(minWidth)
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***));
            addComponents([
                {
                    ".container": Object.assign({
                        width: "100%"
                    ***REMOVED***, theme("container.center", false) ? {
                        marginRight: "auto",
                        marginLeft: "auto"
                    ***REMOVED*** : {***REMOVED***, generatePaddingFor(0))
                ***REMOVED***,
                ...atRules
            ]);
        ***REMOVED***;
    ***REMOVED***)(),
    accessibility: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".sr-only": {
                position: "absolute",
                width: "1px",
                height: "1px",
                padding: "0",
                margin: "-1px",
                overflow: "hidden",
                clip: "rect(0, 0, 0, 0)",
                whiteSpace: "nowrap",
                borderWidth: "0"
            ***REMOVED***,
            ".not-sr-only": {
                position: "static",
                width: "auto",
                height: "auto",
                padding: "0",
                margin: "0",
                overflow: "visible",
                clip: "auto",
                whiteSpace: "normal"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    pointerEvents: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".pointer-events-none": {
                "pointer-events": "none"
            ***REMOVED***,
            ".pointer-events-auto": {
                "pointer-events": "auto"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    visibility: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".visible": {
                visibility: "visible"
            ***REMOVED***,
            ".invisible": {
                visibility: "hidden"
            ***REMOVED***,
            ".collapse": {
                visibility: "collapse"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    position: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".static": {
                position: "static"
            ***REMOVED***,
            ".fixed": {
                position: "fixed"
            ***REMOVED***,
            ".absolute": {
                position: "absolute"
            ***REMOVED***,
            ".relative": {
                position: "relative"
            ***REMOVED***,
            ".sticky": {
                position: "sticky"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    inset: (0, _createUtilityPlugin.default)("inset", [
        [
            "inset",
            [
                "inset"
            ]
        ],
        [
            [
                "inset-x",
                [
                    "left",
                    "right"
                ]
            ],
            [
                "inset-y",
                [
                    "top",
                    "bottom"
                ]
            ]
        ],
        [
            [
                "start",
                [
                    "inset-inline-start"
                ]
            ],
            [
                "end",
                [
                    "inset-inline-end"
                ]
            ],
            [
                "top",
                [
                    "top"
                ]
            ],
            [
                "right",
                [
                    "right"
                ]
            ],
            [
                "bottom",
                [
                    "bottom"
                ]
            ],
            [
                "left",
                [
                    "left"
                ]
            ]
        ]
    ], {
        supportsNegativeValues: true
    ***REMOVED***),
    isolation: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".isolate": {
                isolation: "isolate"
            ***REMOVED***,
            ".isolation-auto": {
                isolation: "auto"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    zIndex: (0, _createUtilityPlugin.default)("zIndex", [
        [
            "z",
            [
                "zIndex"
            ]
        ]
    ], {
        supportsNegativeValues: true
    ***REMOVED***),
    order: (0, _createUtilityPlugin.default)("order", undefined, {
        supportsNegativeValues: true
    ***REMOVED***),
    gridColumn: (0, _createUtilityPlugin.default)("gridColumn", [
        [
            "col",
            [
                "gridColumn"
            ]
        ]
    ]),
    gridColumnStart: (0, _createUtilityPlugin.default)("gridColumnStart", [
        [
            "col-start",
            [
                "gridColumnStart"
            ]
        ]
    ], {
        supportsNegativeValues: true
    ***REMOVED***),
    gridColumnEnd: (0, _createUtilityPlugin.default)("gridColumnEnd", [
        [
            "col-end",
            [
                "gridColumnEnd"
            ]
        ]
    ], {
        supportsNegativeValues: true
    ***REMOVED***),
    gridRow: (0, _createUtilityPlugin.default)("gridRow", [
        [
            "row",
            [
                "gridRow"
            ]
        ]
    ]),
    gridRowStart: (0, _createUtilityPlugin.default)("gridRowStart", [
        [
            "row-start",
            [
                "gridRowStart"
            ]
        ]
    ], {
        supportsNegativeValues: true
    ***REMOVED***),
    gridRowEnd: (0, _createUtilityPlugin.default)("gridRowEnd", [
        [
            "row-end",
            [
                "gridRowEnd"
            ]
        ]
    ], {
        supportsNegativeValues: true
    ***REMOVED***),
    float: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".float-start": {
                float: "inline-start"
            ***REMOVED***,
            ".float-end": {
                float: "inline-end"
            ***REMOVED***,
            ".float-right": {
                float: "right"
            ***REMOVED***,
            ".float-left": {
                float: "left"
            ***REMOVED***,
            ".float-none": {
                float: "none"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    clear: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".clear-start": {
                clear: "inline-start"
            ***REMOVED***,
            ".clear-end": {
                clear: "inline-end"
            ***REMOVED***,
            ".clear-left": {
                clear: "left"
            ***REMOVED***,
            ".clear-right": {
                clear: "right"
            ***REMOVED***,
            ".clear-both": {
                clear: "both"
            ***REMOVED***,
            ".clear-none": {
                clear: "none"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    margin: (0, _createUtilityPlugin.default)("margin", [
        [
            "m",
            [
                "margin"
            ]
        ],
        [
            [
                "mx",
                [
                    "margin-left",
                    "margin-right"
                ]
            ],
            [
                "my",
                [
                    "margin-top",
                    "margin-bottom"
                ]
            ]
        ],
        [
            [
                "ms",
                [
                    "margin-inline-start"
                ]
            ],
            [
                "me",
                [
                    "margin-inline-end"
                ]
            ],
            [
                "mt",
                [
                    "margin-top"
                ]
            ],
            [
                "mr",
                [
                    "margin-right"
                ]
            ],
            [
                "mb",
                [
                    "margin-bottom"
                ]
            ],
            [
                "ml",
                [
                    "margin-left"
                ]
            ]
        ]
    ], {
        supportsNegativeValues: true
    ***REMOVED***),
    boxSizing: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".box-border": {
                "box-sizing": "border-box"
            ***REMOVED***,
            ".box-content": {
                "box-sizing": "content-box"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    lineClamp: ({ matchUtilities , addUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            "line-clamp": (value)=>({
                    overflow: "hidden",
                    display: "-webkit-box",
                    "-webkit-box-orient": "vertical",
                    "-webkit-line-clamp": `${value***REMOVED***`
                ***REMOVED***)
        ***REMOVED***, {
            values: theme("lineClamp")
        ***REMOVED***);
        addUtilities({
            ".line-clamp-none": {
                overflow: "visible",
                display: "block",
                "-webkit-box-orient": "horizontal",
                "-webkit-line-clamp": "none"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    display: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".block": {
                display: "block"
            ***REMOVED***,
            ".inline-block": {
                display: "inline-block"
            ***REMOVED***,
            ".inline": {
                display: "inline"
            ***REMOVED***,
            ".flex": {
                display: "flex"
            ***REMOVED***,
            ".inline-flex": {
                display: "inline-flex"
            ***REMOVED***,
            ".table": {
                display: "table"
            ***REMOVED***,
            ".inline-table": {
                display: "inline-table"
            ***REMOVED***,
            ".table-caption": {
                display: "table-caption"
            ***REMOVED***,
            ".table-cell": {
                display: "table-cell"
            ***REMOVED***,
            ".table-column": {
                display: "table-column"
            ***REMOVED***,
            ".table-column-group": {
                display: "table-column-group"
            ***REMOVED***,
            ".table-footer-group": {
                display: "table-footer-group"
            ***REMOVED***,
            ".table-header-group": {
                display: "table-header-group"
            ***REMOVED***,
            ".table-row-group": {
                display: "table-row-group"
            ***REMOVED***,
            ".table-row": {
                display: "table-row"
            ***REMOVED***,
            ".flow-root": {
                display: "flow-root"
            ***REMOVED***,
            ".grid": {
                display: "grid"
            ***REMOVED***,
            ".inline-grid": {
                display: "inline-grid"
            ***REMOVED***,
            ".contents": {
                display: "contents"
            ***REMOVED***,
            ".list-item": {
                display: "list-item"
            ***REMOVED***,
            ".hidden": {
                display: "none"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    aspectRatio: (0, _createUtilityPlugin.default)("aspectRatio", [
        [
            "aspect",
            [
                "aspect-ratio"
            ]
        ]
    ]),
    size: (0, _createUtilityPlugin.default)("size", [
        [
            "size",
            [
                "width",
                "height"
            ]
        ]
    ]),
    height: (0, _createUtilityPlugin.default)("height", [
        [
            "h",
            [
                "height"
            ]
        ]
    ]),
    maxHeight: (0, _createUtilityPlugin.default)("maxHeight", [
        [
            "max-h",
            [
                "maxHeight"
            ]
        ]
    ]),
    minHeight: (0, _createUtilityPlugin.default)("minHeight", [
        [
            "min-h",
            [
                "minHeight"
            ]
        ]
    ]),
    width: (0, _createUtilityPlugin.default)("width", [
        [
            "w",
            [
                "width"
            ]
        ]
    ]),
    minWidth: (0, _createUtilityPlugin.default)("minWidth", [
        [
            "min-w",
            [
                "minWidth"
            ]
        ]
    ]),
    maxWidth: (0, _createUtilityPlugin.default)("maxWidth", [
        [
            "max-w",
            [
                "maxWidth"
            ]
        ]
    ]),
    flex: (0, _createUtilityPlugin.default)("flex"),
    flexShrink: (0, _createUtilityPlugin.default)("flexShrink", [
        [
            "flex-shrink",
            [
                "flex-shrink"
            ]
        ],
        [
            "shrink",
            [
                "flex-shrink"
            ]
        ]
    ]),
    flexGrow: (0, _createUtilityPlugin.default)("flexGrow", [
        [
            "flex-grow",
            [
                "flex-grow"
            ]
        ],
        [
            "grow",
            [
                "flex-grow"
            ]
        ]
    ]),
    flexBasis: (0, _createUtilityPlugin.default)("flexBasis", [
        [
            "basis",
            [
                "flex-basis"
            ]
        ]
    ]),
    tableLayout: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".table-auto": {
                "table-layout": "auto"
            ***REMOVED***,
            ".table-fixed": {
                "table-layout": "fixed"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    captionSide: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".caption-top": {
                "caption-side": "top"
            ***REMOVED***,
            ".caption-bottom": {
                "caption-side": "bottom"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    borderCollapse: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".border-collapse": {
                "border-collapse": "collapse"
            ***REMOVED***,
            ".border-separate": {
                "border-collapse": "separate"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    borderSpacing: ({ addDefaults , matchUtilities , theme  ***REMOVED***)=>{
        addDefaults("border-spacing", {
            "--tw-border-spacing-x": 0,
            "--tw-border-spacing-y": 0
        ***REMOVED***);
        matchUtilities({
            "border-spacing": (value)=>{
                return {
                    "--tw-border-spacing-x": value,
                    "--tw-border-spacing-y": value,
                    "@defaults border-spacing": {***REMOVED***,
                    "border-spacing": "var(--tw-border-spacing-x) var(--tw-border-spacing-y)"
                ***REMOVED***;
            ***REMOVED***,
            "border-spacing-x": (value)=>{
                return {
                    "--tw-border-spacing-x": value,
                    "@defaults border-spacing": {***REMOVED***,
                    "border-spacing": "var(--tw-border-spacing-x) var(--tw-border-spacing-y)"
                ***REMOVED***;
            ***REMOVED***,
            "border-spacing-y": (value)=>{
                return {
                    "--tw-border-spacing-y": value,
                    "@defaults border-spacing": {***REMOVED***,
                    "border-spacing": "var(--tw-border-spacing-x) var(--tw-border-spacing-y)"
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("borderSpacing")
        ***REMOVED***);
    ***REMOVED***,
    transformOrigin: (0, _createUtilityPlugin.default)("transformOrigin", [
        [
            "origin",
            [
                "transformOrigin"
            ]
        ]
    ]),
    translate: (0, _createUtilityPlugin.default)("translate", [
        [
            [
                "translate-x",
                [
                    [
                        "@defaults transform",
                        {***REMOVED***
                    ],
                    "--tw-translate-x",
                    [
                        "transform",
                        cssTransformValue
                    ]
                ]
            ],
            [
                "translate-y",
                [
                    [
                        "@defaults transform",
                        {***REMOVED***
                    ],
                    "--tw-translate-y",
                    [
                        "transform",
                        cssTransformValue
                    ]
                ]
            ]
        ]
    ], {
        supportsNegativeValues: true
    ***REMOVED***),
    rotate: (0, _createUtilityPlugin.default)("rotate", [
        [
            "rotate",
            [
                [
                    "@defaults transform",
                    {***REMOVED***
                ],
                "--tw-rotate",
                [
                    "transform",
                    cssTransformValue
                ]
            ]
        ]
    ], {
        supportsNegativeValues: true
    ***REMOVED***),
    skew: (0, _createUtilityPlugin.default)("skew", [
        [
            [
                "skew-x",
                [
                    [
                        "@defaults transform",
                        {***REMOVED***
                    ],
                    "--tw-skew-x",
                    [
                        "transform",
                        cssTransformValue
                    ]
                ]
            ],
            [
                "skew-y",
                [
                    [
                        "@defaults transform",
                        {***REMOVED***
                    ],
                    "--tw-skew-y",
                    [
                        "transform",
                        cssTransformValue
                    ]
                ]
            ]
        ]
    ], {
        supportsNegativeValues: true
    ***REMOVED***),
    scale: (0, _createUtilityPlugin.default)("scale", [
        [
            "scale",
            [
                [
                    "@defaults transform",
                    {***REMOVED***
                ],
                "--tw-scale-x",
                "--tw-scale-y",
                [
                    "transform",
                    cssTransformValue
                ]
            ]
        ],
        [
            [
                "scale-x",
                [
                    [
                        "@defaults transform",
                        {***REMOVED***
                    ],
                    "--tw-scale-x",
                    [
                        "transform",
                        cssTransformValue
                    ]
                ]
            ],
            [
                "scale-y",
                [
                    [
                        "@defaults transform",
                        {***REMOVED***
                    ],
                    "--tw-scale-y",
                    [
                        "transform",
                        cssTransformValue
                    ]
                ]
            ]
        ]
    ], {
        supportsNegativeValues: true
    ***REMOVED***),
    transform: ({ addDefaults , addUtilities  ***REMOVED***)=>{
        addDefaults("transform", {
            "--tw-translate-x": "0",
            "--tw-translate-y": "0",
            "--tw-rotate": "0",
            "--tw-skew-x": "0",
            "--tw-skew-y": "0",
            "--tw-scale-x": "1",
            "--tw-scale-y": "1"
        ***REMOVED***);
        addUtilities({
            ".transform": {
                "@defaults transform": {***REMOVED***,
                transform: cssTransformValue
            ***REMOVED***,
            ".transform-cpu": {
                transform: cssTransformValue
            ***REMOVED***,
            ".transform-gpu": {
                transform: cssTransformValue.replace("translate(var(--tw-translate-x), var(--tw-translate-y))", "translate3d(var(--tw-translate-x), var(--tw-translate-y), 0)")
            ***REMOVED***,
            ".transform-none": {
                transform: "none"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    animation: ({ matchUtilities , theme , config  ***REMOVED***)=>{
        let prefixName = (name)=>(0, _escapeClassName.default)(config("prefix") + name);
        var _theme;
        let keyframes = Object.fromEntries(Object.entries((_theme = theme("keyframes")) !== null && _theme !== void 0 ? _theme : {***REMOVED***).map(([key, value])=>{
            return [
                key,
                {
                    [`@keyframes ${prefixName(key)***REMOVED***`]: value
                ***REMOVED***
            ];
        ***REMOVED***));
        matchUtilities({
            animate: (value)=>{
                let animations = (0, _parseAnimationValue.default)(value);
                return [
                    ...animations.flatMap((animation)=>keyframes[animation.name]),
                    {
                        animation: animations.map(({ name , value  ***REMOVED***)=>{
                            if (name === undefined || keyframes[name] === undefined) {
                                return value;
                            ***REMOVED***
                            return value.replace(name, prefixName(name));
                        ***REMOVED***).join(", ")
                    ***REMOVED***
                ];
            ***REMOVED***
        ***REMOVED***, {
            values: theme("animation")
        ***REMOVED***);
    ***REMOVED***,
    cursor: (0, _createUtilityPlugin.default)("cursor"),
    touchAction: ({ addDefaults , addUtilities  ***REMOVED***)=>{
        addDefaults("touch-action", {
            "--tw-pan-x": " ",
            "--tw-pan-y": " ",
            "--tw-pinch-zoom": " "
        ***REMOVED***);
        let cssTouchActionValue = "var(--tw-pan-x) var(--tw-pan-y) var(--tw-pinch-zoom)";
        addUtilities({
            ".touch-auto": {
                "touch-action": "auto"
            ***REMOVED***,
            ".touch-none": {
                "touch-action": "none"
            ***REMOVED***,
            ".touch-pan-x": {
                "@defaults touch-action": {***REMOVED***,
                "--tw-pan-x": "pan-x",
                "touch-action": cssTouchActionValue
            ***REMOVED***,
            ".touch-pan-left": {
                "@defaults touch-action": {***REMOVED***,
                "--tw-pan-x": "pan-left",
                "touch-action": cssTouchActionValue
            ***REMOVED***,
            ".touch-pan-right": {
                "@defaults touch-action": {***REMOVED***,
                "--tw-pan-x": "pan-right",
                "touch-action": cssTouchActionValue
            ***REMOVED***,
            ".touch-pan-y": {
                "@defaults touch-action": {***REMOVED***,
                "--tw-pan-y": "pan-y",
                "touch-action": cssTouchActionValue
            ***REMOVED***,
            ".touch-pan-up": {
                "@defaults touch-action": {***REMOVED***,
                "--tw-pan-y": "pan-up",
                "touch-action": cssTouchActionValue
            ***REMOVED***,
            ".touch-pan-down": {
                "@defaults touch-action": {***REMOVED***,
                "--tw-pan-y": "pan-down",
                "touch-action": cssTouchActionValue
            ***REMOVED***,
            ".touch-pinch-zoom": {
                "@defaults touch-action": {***REMOVED***,
                "--tw-pinch-zoom": "pinch-zoom",
                "touch-action": cssTouchActionValue
            ***REMOVED***,
            ".touch-manipulation": {
                "touch-action": "manipulation"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    userSelect: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".select-none": {
                "user-select": "none"
            ***REMOVED***,
            ".select-text": {
                "user-select": "text"
            ***REMOVED***,
            ".select-all": {
                "user-select": "all"
            ***REMOVED***,
            ".select-auto": {
                "user-select": "auto"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    resize: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".resize-none": {
                resize: "none"
            ***REMOVED***,
            ".resize-y": {
                resize: "vertical"
            ***REMOVED***,
            ".resize-x": {
                resize: "horizontal"
            ***REMOVED***,
            ".resize": {
                resize: "both"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    scrollSnapType: ({ addDefaults , addUtilities  ***REMOVED***)=>{
        addDefaults("scroll-snap-type", {
            "--tw-scroll-snap-strictness": "proximity"
        ***REMOVED***);
        addUtilities({
            ".snap-none": {
                "scroll-snap-type": "none"
            ***REMOVED***,
            ".snap-x": {
                "@defaults scroll-snap-type": {***REMOVED***,
                "scroll-snap-type": "x var(--tw-scroll-snap-strictness)"
            ***REMOVED***,
            ".snap-y": {
                "@defaults scroll-snap-type": {***REMOVED***,
                "scroll-snap-type": "y var(--tw-scroll-snap-strictness)"
            ***REMOVED***,
            ".snap-both": {
                "@defaults scroll-snap-type": {***REMOVED***,
                "scroll-snap-type": "both var(--tw-scroll-snap-strictness)"
            ***REMOVED***,
            ".snap-mandatory": {
                "--tw-scroll-snap-strictness": "mandatory"
            ***REMOVED***,
            ".snap-proximity": {
                "--tw-scroll-snap-strictness": "proximity"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    scrollSnapAlign: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".snap-start": {
                "scroll-snap-align": "start"
            ***REMOVED***,
            ".snap-end": {
                "scroll-snap-align": "end"
            ***REMOVED***,
            ".snap-center": {
                "scroll-snap-align": "center"
            ***REMOVED***,
            ".snap-align-none": {
                "scroll-snap-align": "none"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    scrollSnapStop: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".snap-normal": {
                "scroll-snap-stop": "normal"
            ***REMOVED***,
            ".snap-always": {
                "scroll-snap-stop": "always"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    scrollMargin: (0, _createUtilityPlugin.default)("scrollMargin", [
        [
            "scroll-m",
            [
                "scroll-margin"
            ]
        ],
        [
            [
                "scroll-mx",
                [
                    "scroll-margin-left",
                    "scroll-margin-right"
                ]
            ],
            [
                "scroll-my",
                [
                    "scroll-margin-top",
                    "scroll-margin-bottom"
                ]
            ]
        ],
        [
            [
                "scroll-ms",
                [
                    "scroll-margin-inline-start"
                ]
            ],
            [
                "scroll-me",
                [
                    "scroll-margin-inline-end"
                ]
            ],
            [
                "scroll-mt",
                [
                    "scroll-margin-top"
                ]
            ],
            [
                "scroll-mr",
                [
                    "scroll-margin-right"
                ]
            ],
            [
                "scroll-mb",
                [
                    "scroll-margin-bottom"
                ]
            ],
            [
                "scroll-ml",
                [
                    "scroll-margin-left"
                ]
            ]
        ]
    ], {
        supportsNegativeValues: true
    ***REMOVED***),
    scrollPadding: (0, _createUtilityPlugin.default)("scrollPadding", [
        [
            "scroll-p",
            [
                "scroll-padding"
            ]
        ],
        [
            [
                "scroll-px",
                [
                    "scroll-padding-left",
                    "scroll-padding-right"
                ]
            ],
            [
                "scroll-py",
                [
                    "scroll-padding-top",
                    "scroll-padding-bottom"
                ]
            ]
        ],
        [
            [
                "scroll-ps",
                [
                    "scroll-padding-inline-start"
                ]
            ],
            [
                "scroll-pe",
                [
                    "scroll-padding-inline-end"
                ]
            ],
            [
                "scroll-pt",
                [
                    "scroll-padding-top"
                ]
            ],
            [
                "scroll-pr",
                [
                    "scroll-padding-right"
                ]
            ],
            [
                "scroll-pb",
                [
                    "scroll-padding-bottom"
                ]
            ],
            [
                "scroll-pl",
                [
                    "scroll-padding-left"
                ]
            ]
        ]
    ]),
    listStylePosition: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".list-inside": {
                "list-style-position": "inside"
            ***REMOVED***,
            ".list-outside": {
                "list-style-position": "outside"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    listStyleType: (0, _createUtilityPlugin.default)("listStyleType", [
        [
            "list",
            [
                "listStyleType"
            ]
        ]
    ]),
    listStyleImage: (0, _createUtilityPlugin.default)("listStyleImage", [
        [
            "list-image",
            [
                "listStyleImage"
            ]
        ]
    ]),
    appearance: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".appearance-none": {
                appearance: "none"
            ***REMOVED***,
            ".appearance-auto": {
                appearance: "auto"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    columns: (0, _createUtilityPlugin.default)("columns", [
        [
            "columns",
            [
                "columns"
            ]
        ]
    ]),
    breakBefore: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".break-before-auto": {
                "break-before": "auto"
            ***REMOVED***,
            ".break-before-avoid": {
                "break-before": "avoid"
            ***REMOVED***,
            ".break-before-all": {
                "break-before": "all"
            ***REMOVED***,
            ".break-before-avoid-page": {
                "break-before": "avoid-page"
            ***REMOVED***,
            ".break-before-page": {
                "break-before": "page"
            ***REMOVED***,
            ".break-before-left": {
                "break-before": "left"
            ***REMOVED***,
            ".break-before-right": {
                "break-before": "right"
            ***REMOVED***,
            ".break-before-column": {
                "break-before": "column"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    breakInside: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".break-inside-auto": {
                "break-inside": "auto"
            ***REMOVED***,
            ".break-inside-avoid": {
                "break-inside": "avoid"
            ***REMOVED***,
            ".break-inside-avoid-page": {
                "break-inside": "avoid-page"
            ***REMOVED***,
            ".break-inside-avoid-column": {
                "break-inside": "avoid-column"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    breakAfter: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".break-after-auto": {
                "break-after": "auto"
            ***REMOVED***,
            ".break-after-avoid": {
                "break-after": "avoid"
            ***REMOVED***,
            ".break-after-all": {
                "break-after": "all"
            ***REMOVED***,
            ".break-after-avoid-page": {
                "break-after": "avoid-page"
            ***REMOVED***,
            ".break-after-page": {
                "break-after": "page"
            ***REMOVED***,
            ".break-after-left": {
                "break-after": "left"
            ***REMOVED***,
            ".break-after-right": {
                "break-after": "right"
            ***REMOVED***,
            ".break-after-column": {
                "break-after": "column"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    gridAutoColumns: (0, _createUtilityPlugin.default)("gridAutoColumns", [
        [
            "auto-cols",
            [
                "gridAutoColumns"
            ]
        ]
    ]),
    gridAutoFlow: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".grid-flow-row": {
                gridAutoFlow: "row"
            ***REMOVED***,
            ".grid-flow-col": {
                gridAutoFlow: "column"
            ***REMOVED***,
            ".grid-flow-dense": {
                gridAutoFlow: "dense"
            ***REMOVED***,
            ".grid-flow-row-dense": {
                gridAutoFlow: "row dense"
            ***REMOVED***,
            ".grid-flow-col-dense": {
                gridAutoFlow: "column dense"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    gridAutoRows: (0, _createUtilityPlugin.default)("gridAutoRows", [
        [
            "auto-rows",
            [
                "gridAutoRows"
            ]
        ]
    ]),
    gridTemplateColumns: (0, _createUtilityPlugin.default)("gridTemplateColumns", [
        [
            "grid-cols",
            [
                "gridTemplateColumns"
            ]
        ]
    ]),
    gridTemplateRows: (0, _createUtilityPlugin.default)("gridTemplateRows", [
        [
            "grid-rows",
            [
                "gridTemplateRows"
            ]
        ]
    ]),
    flexDirection: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".flex-row": {
                "flex-direction": "row"
            ***REMOVED***,
            ".flex-row-reverse": {
                "flex-direction": "row-reverse"
            ***REMOVED***,
            ".flex-col": {
                "flex-direction": "column"
            ***REMOVED***,
            ".flex-col-reverse": {
                "flex-direction": "column-reverse"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    flexWrap: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".flex-wrap": {
                "flex-wrap": "wrap"
            ***REMOVED***,
            ".flex-wrap-reverse": {
                "flex-wrap": "wrap-reverse"
            ***REMOVED***,
            ".flex-nowrap": {
                "flex-wrap": "nowrap"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    placeContent: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".place-content-center": {
                "place-content": "center"
            ***REMOVED***,
            ".place-content-start": {
                "place-content": "start"
            ***REMOVED***,
            ".place-content-end": {
                "place-content": "end"
            ***REMOVED***,
            ".place-content-between": {
                "place-content": "space-between"
            ***REMOVED***,
            ".place-content-around": {
                "place-content": "space-around"
            ***REMOVED***,
            ".place-content-evenly": {
                "place-content": "space-evenly"
            ***REMOVED***,
            ".place-content-baseline": {
                "place-content": "baseline"
            ***REMOVED***,
            ".place-content-stretch": {
                "place-content": "stretch"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    placeItems: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".place-items-start": {
                "place-items": "start"
            ***REMOVED***,
            ".place-items-end": {
                "place-items": "end"
            ***REMOVED***,
            ".place-items-center": {
                "place-items": "center"
            ***REMOVED***,
            ".place-items-baseline": {
                "place-items": "baseline"
            ***REMOVED***,
            ".place-items-stretch": {
                "place-items": "stretch"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    alignContent: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".content-normal": {
                "align-content": "normal"
            ***REMOVED***,
            ".content-center": {
                "align-content": "center"
            ***REMOVED***,
            ".content-start": {
                "align-content": "flex-start"
            ***REMOVED***,
            ".content-end": {
                "align-content": "flex-end"
            ***REMOVED***,
            ".content-between": {
                "align-content": "space-between"
            ***REMOVED***,
            ".content-around": {
                "align-content": "space-around"
            ***REMOVED***,
            ".content-evenly": {
                "align-content": "space-evenly"
            ***REMOVED***,
            ".content-baseline": {
                "align-content": "baseline"
            ***REMOVED***,
            ".content-stretch": {
                "align-content": "stretch"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    alignItems: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".items-start": {
                "align-items": "flex-start"
            ***REMOVED***,
            ".items-end": {
                "align-items": "flex-end"
            ***REMOVED***,
            ".items-center": {
                "align-items": "center"
            ***REMOVED***,
            ".items-baseline": {
                "align-items": "baseline"
            ***REMOVED***,
            ".items-stretch": {
                "align-items": "stretch"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    justifyContent: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".justify-normal": {
                "justify-content": "normal"
            ***REMOVED***,
            ".justify-start": {
                "justify-content": "flex-start"
            ***REMOVED***,
            ".justify-end": {
                "justify-content": "flex-end"
            ***REMOVED***,
            ".justify-center": {
                "justify-content": "center"
            ***REMOVED***,
            ".justify-between": {
                "justify-content": "space-between"
            ***REMOVED***,
            ".justify-around": {
                "justify-content": "space-around"
            ***REMOVED***,
            ".justify-evenly": {
                "justify-content": "space-evenly"
            ***REMOVED***,
            ".justify-stretch": {
                "justify-content": "stretch"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    justifyItems: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".justify-items-start": {
                "justify-items": "start"
            ***REMOVED***,
            ".justify-items-end": {
                "justify-items": "end"
            ***REMOVED***,
            ".justify-items-center": {
                "justify-items": "center"
            ***REMOVED***,
            ".justify-items-stretch": {
                "justify-items": "stretch"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    gap: (0, _createUtilityPlugin.default)("gap", [
        [
            "gap",
            [
                "gap"
            ]
        ],
        [
            [
                "gap-x",
                [
                    "columnGap"
                ]
            ],
            [
                "gap-y",
                [
                    "rowGap"
                ]
            ]
        ]
    ]),
    space: ({ matchUtilities , addUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            "space-x": (value)=>{
                value = value === "0" ? "0px" : value;
                return {
                    "& > :not([hidden]) ~ :not([hidden])": {
                        "--tw-space-x-reverse": "0",
                        "margin-right": `calc(${value***REMOVED*** * var(--tw-space-x-reverse))`,
                        "margin-left": `calc(${value***REMOVED*** * calc(1 - var(--tw-space-x-reverse)))`
                    ***REMOVED***
                ***REMOVED***;
            ***REMOVED***,
            "space-y": (value)=>{
                value = value === "0" ? "0px" : value;
                return {
                    "& > :not([hidden]) ~ :not([hidden])": {
                        "--tw-space-y-reverse": "0",
                        "margin-top": `calc(${value***REMOVED*** * calc(1 - var(--tw-space-y-reverse)))`,
                        "margin-bottom": `calc(${value***REMOVED*** * var(--tw-space-y-reverse))`
                    ***REMOVED***
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("space"),
            supportsNegativeValues: true
        ***REMOVED***);
        addUtilities({
            ".space-y-reverse > :not([hidden]) ~ :not([hidden])": {
                "--tw-space-y-reverse": "1"
            ***REMOVED***,
            ".space-x-reverse > :not([hidden]) ~ :not([hidden])": {
                "--tw-space-x-reverse": "1"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    divideWidth: ({ matchUtilities , addUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            "divide-x": (value)=>{
                value = value === "0" ? "0px" : value;
                return {
                    "& > :not([hidden]) ~ :not([hidden])": {
                        "@defaults border-width": {***REMOVED***,
                        "--tw-divide-x-reverse": "0",
                        "border-right-width": `calc(${value***REMOVED*** * var(--tw-divide-x-reverse))`,
                        "border-left-width": `calc(${value***REMOVED*** * calc(1 - var(--tw-divide-x-reverse)))`
                    ***REMOVED***
                ***REMOVED***;
            ***REMOVED***,
            "divide-y": (value)=>{
                value = value === "0" ? "0px" : value;
                return {
                    "& > :not([hidden]) ~ :not([hidden])": {
                        "@defaults border-width": {***REMOVED***,
                        "--tw-divide-y-reverse": "0",
                        "border-top-width": `calc(${value***REMOVED*** * calc(1 - var(--tw-divide-y-reverse)))`,
                        "border-bottom-width": `calc(${value***REMOVED*** * var(--tw-divide-y-reverse))`
                    ***REMOVED***
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("divideWidth"),
            type: [
                "line-width",
                "length",
                "any"
            ]
        ***REMOVED***);
        addUtilities({
            ".divide-y-reverse > :not([hidden]) ~ :not([hidden])": {
                "@defaults border-width": {***REMOVED***,
                "--tw-divide-y-reverse": "1"
            ***REMOVED***,
            ".divide-x-reverse > :not([hidden]) ~ :not([hidden])": {
                "@defaults border-width": {***REMOVED***,
                "--tw-divide-x-reverse": "1"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    divideStyle: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".divide-solid > :not([hidden]) ~ :not([hidden])": {
                "border-style": "solid"
            ***REMOVED***,
            ".divide-dashed > :not([hidden]) ~ :not([hidden])": {
                "border-style": "dashed"
            ***REMOVED***,
            ".divide-dotted > :not([hidden]) ~ :not([hidden])": {
                "border-style": "dotted"
            ***REMOVED***,
            ".divide-double > :not([hidden]) ~ :not([hidden])": {
                "border-style": "double"
            ***REMOVED***,
            ".divide-none > :not([hidden]) ~ :not([hidden])": {
                "border-style": "none"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    divideColor: ({ matchUtilities , theme , corePlugins  ***REMOVED***)=>{
        matchUtilities({
            divide: (value)=>{
                if (!corePlugins("divideOpacity")) {
                    return {
                        ["& > :not([hidden]) ~ :not([hidden])"]: {
                            "border-color": (0, _toColorValue.default)(value)
                        ***REMOVED***
                    ***REMOVED***;
                ***REMOVED***
                return {
                    ["& > :not([hidden]) ~ :not([hidden])"]: (0, _withAlphaVariable.default)({
                        color: value,
                        property: "border-color",
                        variable: "--tw-divide-opacity"
                    ***REMOVED***)
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: (({ DEFAULT: _ , ...colors ***REMOVED***)=>colors)((0, _flattenColorPalette.default)(theme("divideColor"))),
            type: [
                "color",
                "any"
            ]
        ***REMOVED***);
    ***REMOVED***,
    divideOpacity: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            "divide-opacity": (value)=>{
                return {
                    [`& > :not([hidden]) ~ :not([hidden])`]: {
                        "--tw-divide-opacity": value
                    ***REMOVED***
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("divideOpacity")
        ***REMOVED***);
    ***REMOVED***,
    placeSelf: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".place-self-auto": {
                "place-self": "auto"
            ***REMOVED***,
            ".place-self-start": {
                "place-self": "start"
            ***REMOVED***,
            ".place-self-end": {
                "place-self": "end"
            ***REMOVED***,
            ".place-self-center": {
                "place-self": "center"
            ***REMOVED***,
            ".place-self-stretch": {
                "place-self": "stretch"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    alignSelf: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".self-auto": {
                "align-self": "auto"
            ***REMOVED***,
            ".self-start": {
                "align-self": "flex-start"
            ***REMOVED***,
            ".self-end": {
                "align-self": "flex-end"
            ***REMOVED***,
            ".self-center": {
                "align-self": "center"
            ***REMOVED***,
            ".self-stretch": {
                "align-self": "stretch"
            ***REMOVED***,
            ".self-baseline": {
                "align-self": "baseline"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    justifySelf: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".justify-self-auto": {
                "justify-self": "auto"
            ***REMOVED***,
            ".justify-self-start": {
                "justify-self": "start"
            ***REMOVED***,
            ".justify-self-end": {
                "justify-self": "end"
            ***REMOVED***,
            ".justify-self-center": {
                "justify-self": "center"
            ***REMOVED***,
            ".justify-self-stretch": {
                "justify-self": "stretch"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    overflow: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".overflow-auto": {
                overflow: "auto"
            ***REMOVED***,
            ".overflow-hidden": {
                overflow: "hidden"
            ***REMOVED***,
            ".overflow-clip": {
                overflow: "clip"
            ***REMOVED***,
            ".overflow-visible": {
                overflow: "visible"
            ***REMOVED***,
            ".overflow-scroll": {
                overflow: "scroll"
            ***REMOVED***,
            ".overflow-x-auto": {
                "overflow-x": "auto"
            ***REMOVED***,
            ".overflow-y-auto": {
                "overflow-y": "auto"
            ***REMOVED***,
            ".overflow-x-hidden": {
                "overflow-x": "hidden"
            ***REMOVED***,
            ".overflow-y-hidden": {
                "overflow-y": "hidden"
            ***REMOVED***,
            ".overflow-x-clip": {
                "overflow-x": "clip"
            ***REMOVED***,
            ".overflow-y-clip": {
                "overflow-y": "clip"
            ***REMOVED***,
            ".overflow-x-visible": {
                "overflow-x": "visible"
            ***REMOVED***,
            ".overflow-y-visible": {
                "overflow-y": "visible"
            ***REMOVED***,
            ".overflow-x-scroll": {
                "overflow-x": "scroll"
            ***REMOVED***,
            ".overflow-y-scroll": {
                "overflow-y": "scroll"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    overscrollBehavior: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".overscroll-auto": {
                "overscroll-behavior": "auto"
            ***REMOVED***,
            ".overscroll-contain": {
                "overscroll-behavior": "contain"
            ***REMOVED***,
            ".overscroll-none": {
                "overscroll-behavior": "none"
            ***REMOVED***,
            ".overscroll-y-auto": {
                "overscroll-behavior-y": "auto"
            ***REMOVED***,
            ".overscroll-y-contain": {
                "overscroll-behavior-y": "contain"
            ***REMOVED***,
            ".overscroll-y-none": {
                "overscroll-behavior-y": "none"
            ***REMOVED***,
            ".overscroll-x-auto": {
                "overscroll-behavior-x": "auto"
            ***REMOVED***,
            ".overscroll-x-contain": {
                "overscroll-behavior-x": "contain"
            ***REMOVED***,
            ".overscroll-x-none": {
                "overscroll-behavior-x": "none"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    scrollBehavior: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".scroll-auto": {
                "scroll-behavior": "auto"
            ***REMOVED***,
            ".scroll-smooth": {
                "scroll-behavior": "smooth"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    textOverflow: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".truncate": {
                overflow: "hidden",
                "text-overflow": "ellipsis",
                "white-space": "nowrap"
            ***REMOVED***,
            ".overflow-ellipsis": {
                "text-overflow": "ellipsis"
            ***REMOVED***,
            ".text-ellipsis": {
                "text-overflow": "ellipsis"
            ***REMOVED***,
            ".text-clip": {
                "text-overflow": "clip"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    hyphens: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".hyphens-none": {
                hyphens: "none"
            ***REMOVED***,
            ".hyphens-manual": {
                hyphens: "manual"
            ***REMOVED***,
            ".hyphens-auto": {
                hyphens: "auto"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    whitespace: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".whitespace-normal": {
                "white-space": "normal"
            ***REMOVED***,
            ".whitespace-nowrap": {
                "white-space": "nowrap"
            ***REMOVED***,
            ".whitespace-pre": {
                "white-space": "pre"
            ***REMOVED***,
            ".whitespace-pre-line": {
                "white-space": "pre-line"
            ***REMOVED***,
            ".whitespace-pre-wrap": {
                "white-space": "pre-wrap"
            ***REMOVED***,
            ".whitespace-break-spaces": {
                "white-space": "break-spaces"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    textWrap: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".text-wrap": {
                "text-wrap": "wrap"
            ***REMOVED***,
            ".text-nowrap": {
                "text-wrap": "nowrap"
            ***REMOVED***,
            ".text-balance": {
                "text-wrap": "balance"
            ***REMOVED***,
            ".text-pretty": {
                "text-wrap": "pretty"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    wordBreak: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".break-normal": {
                "overflow-wrap": "normal",
                "word-break": "normal"
            ***REMOVED***,
            ".break-words": {
                "overflow-wrap": "break-word"
            ***REMOVED***,
            ".break-all": {
                "word-break": "break-all"
            ***REMOVED***,
            ".break-keep": {
                "word-break": "keep-all"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    borderRadius: (0, _createUtilityPlugin.default)("borderRadius", [
        [
            "rounded",
            [
                "border-radius"
            ]
        ],
        [
            [
                "rounded-s",
                [
                    "border-start-start-radius",
                    "border-end-start-radius"
                ]
            ],
            [
                "rounded-e",
                [
                    "border-start-end-radius",
                    "border-end-end-radius"
                ]
            ],
            [
                "rounded-t",
                [
                    "border-top-left-radius",
                    "border-top-right-radius"
                ]
            ],
            [
                "rounded-r",
                [
                    "border-top-right-radius",
                    "border-bottom-right-radius"
                ]
            ],
            [
                "rounded-b",
                [
                    "border-bottom-right-radius",
                    "border-bottom-left-radius"
                ]
            ],
            [
                "rounded-l",
                [
                    "border-top-left-radius",
                    "border-bottom-left-radius"
                ]
            ]
        ],
        [
            [
                "rounded-ss",
                [
                    "border-start-start-radius"
                ]
            ],
            [
                "rounded-se",
                [
                    "border-start-end-radius"
                ]
            ],
            [
                "rounded-ee",
                [
                    "border-end-end-radius"
                ]
            ],
            [
                "rounded-es",
                [
                    "border-end-start-radius"
                ]
            ],
            [
                "rounded-tl",
                [
                    "border-top-left-radius"
                ]
            ],
            [
                "rounded-tr",
                [
                    "border-top-right-radius"
                ]
            ],
            [
                "rounded-br",
                [
                    "border-bottom-right-radius"
                ]
            ],
            [
                "rounded-bl",
                [
                    "border-bottom-left-radius"
                ]
            ]
        ]
    ]),
    borderWidth: (0, _createUtilityPlugin.default)("borderWidth", [
        [
            "border",
            [
                [
                    "@defaults border-width",
                    {***REMOVED***
                ],
                "border-width"
            ]
        ],
        [
            [
                "border-x",
                [
                    [
                        "@defaults border-width",
                        {***REMOVED***
                    ],
                    "border-left-width",
                    "border-right-width"
                ]
            ],
            [
                "border-y",
                [
                    [
                        "@defaults border-width",
                        {***REMOVED***
                    ],
                    "border-top-width",
                    "border-bottom-width"
                ]
            ]
        ],
        [
            [
                "border-s",
                [
                    [
                        "@defaults border-width",
                        {***REMOVED***
                    ],
                    "border-inline-start-width"
                ]
            ],
            [
                "border-e",
                [
                    [
                        "@defaults border-width",
                        {***REMOVED***
                    ],
                    "border-inline-end-width"
                ]
            ],
            [
                "border-t",
                [
                    [
                        "@defaults border-width",
                        {***REMOVED***
                    ],
                    "border-top-width"
                ]
            ],
            [
                "border-r",
                [
                    [
                        "@defaults border-width",
                        {***REMOVED***
                    ],
                    "border-right-width"
                ]
            ],
            [
                "border-b",
                [
                    [
                        "@defaults border-width",
                        {***REMOVED***
                    ],
                    "border-bottom-width"
                ]
            ],
            [
                "border-l",
                [
                    [
                        "@defaults border-width",
                        {***REMOVED***
                    ],
                    "border-left-width"
                ]
            ]
        ]
    ], {
        type: [
            "line-width",
            "length"
        ]
    ***REMOVED***),
    borderStyle: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".border-solid": {
                "border-style": "solid"
            ***REMOVED***,
            ".border-dashed": {
                "border-style": "dashed"
            ***REMOVED***,
            ".border-dotted": {
                "border-style": "dotted"
            ***REMOVED***,
            ".border-double": {
                "border-style": "double"
            ***REMOVED***,
            ".border-hidden": {
                "border-style": "hidden"
            ***REMOVED***,
            ".border-none": {
                "border-style": "none"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    borderColor: ({ matchUtilities , theme , corePlugins  ***REMOVED***)=>{
        matchUtilities({
            border: (value)=>{
                if (!corePlugins("borderOpacity")) {
                    return {
                        "border-color": (0, _toColorValue.default)(value)
                    ***REMOVED***;
                ***REMOVED***
                return (0, _withAlphaVariable.default)({
                    color: value,
                    property: "border-color",
                    variable: "--tw-border-opacity"
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***, {
            values: (({ DEFAULT: _ , ...colors ***REMOVED***)=>colors)((0, _flattenColorPalette.default)(theme("borderColor"))),
            type: [
                "color",
                "any"
            ]
        ***REMOVED***);
        matchUtilities({
            "border-x": (value)=>{
                if (!corePlugins("borderOpacity")) {
                    return {
                        "border-left-color": (0, _toColorValue.default)(value),
                        "border-right-color": (0, _toColorValue.default)(value)
                    ***REMOVED***;
                ***REMOVED***
                return (0, _withAlphaVariable.default)({
                    color: value,
                    property: [
                        "border-left-color",
                        "border-right-color"
                    ],
                    variable: "--tw-border-opacity"
                ***REMOVED***);
            ***REMOVED***,
            "border-y": (value)=>{
                if (!corePlugins("borderOpacity")) {
                    return {
                        "border-top-color": (0, _toColorValue.default)(value),
                        "border-bottom-color": (0, _toColorValue.default)(value)
                    ***REMOVED***;
                ***REMOVED***
                return (0, _withAlphaVariable.default)({
                    color: value,
                    property: [
                        "border-top-color",
                        "border-bottom-color"
                    ],
                    variable: "--tw-border-opacity"
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***, {
            values: (({ DEFAULT: _ , ...colors ***REMOVED***)=>colors)((0, _flattenColorPalette.default)(theme("borderColor"))),
            type: [
                "color",
                "any"
            ]
        ***REMOVED***);
        matchUtilities({
            "border-s": (value)=>{
                if (!corePlugins("borderOpacity")) {
                    return {
                        "border-inline-start-color": (0, _toColorValue.default)(value)
                    ***REMOVED***;
                ***REMOVED***
                return (0, _withAlphaVariable.default)({
                    color: value,
                    property: "border-inline-start-color",
                    variable: "--tw-border-opacity"
                ***REMOVED***);
            ***REMOVED***,
            "border-e": (value)=>{
                if (!corePlugins("borderOpacity")) {
                    return {
                        "border-inline-end-color": (0, _toColorValue.default)(value)
                    ***REMOVED***;
                ***REMOVED***
                return (0, _withAlphaVariable.default)({
                    color: value,
                    property: "border-inline-end-color",
                    variable: "--tw-border-opacity"
                ***REMOVED***);
            ***REMOVED***,
            "border-t": (value)=>{
                if (!corePlugins("borderOpacity")) {
                    return {
                        "border-top-color": (0, _toColorValue.default)(value)
                    ***REMOVED***;
                ***REMOVED***
                return (0, _withAlphaVariable.default)({
                    color: value,
                    property: "border-top-color",
                    variable: "--tw-border-opacity"
                ***REMOVED***);
            ***REMOVED***,
            "border-r": (value)=>{
                if (!corePlugins("borderOpacity")) {
                    return {
                        "border-right-color": (0, _toColorValue.default)(value)
                    ***REMOVED***;
                ***REMOVED***
                return (0, _withAlphaVariable.default)({
                    color: value,
                    property: "border-right-color",
                    variable: "--tw-border-opacity"
                ***REMOVED***);
            ***REMOVED***,
            "border-b": (value)=>{
                if (!corePlugins("borderOpacity")) {
                    return {
                        "border-bottom-color": (0, _toColorValue.default)(value)
                    ***REMOVED***;
                ***REMOVED***
                return (0, _withAlphaVariable.default)({
                    color: value,
                    property: "border-bottom-color",
                    variable: "--tw-border-opacity"
                ***REMOVED***);
            ***REMOVED***,
            "border-l": (value)=>{
                if (!corePlugins("borderOpacity")) {
                    return {
                        "border-left-color": (0, _toColorValue.default)(value)
                    ***REMOVED***;
                ***REMOVED***
                return (0, _withAlphaVariable.default)({
                    color: value,
                    property: "border-left-color",
                    variable: "--tw-border-opacity"
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***, {
            values: (({ DEFAULT: _ , ...colors ***REMOVED***)=>colors)((0, _flattenColorPalette.default)(theme("borderColor"))),
            type: [
                "color",
                "any"
            ]
        ***REMOVED***);
    ***REMOVED***,
    borderOpacity: (0, _createUtilityPlugin.default)("borderOpacity", [
        [
            "border-opacity",
            [
                "--tw-border-opacity"
            ]
        ]
    ]),
    backgroundColor: ({ matchUtilities , theme , corePlugins  ***REMOVED***)=>{
        matchUtilities({
            bg: (value)=>{
                if (!corePlugins("backgroundOpacity")) {
                    return {
                        "background-color": (0, _toColorValue.default)(value)
                    ***REMOVED***;
                ***REMOVED***
                return (0, _withAlphaVariable.default)({
                    color: value,
                    property: "background-color",
                    variable: "--tw-bg-opacity"
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***, {
            values: (0, _flattenColorPalette.default)(theme("backgroundColor")),
            type: [
                "color",
                "any"
            ]
        ***REMOVED***);
    ***REMOVED***,
    backgroundOpacity: (0, _createUtilityPlugin.default)("backgroundOpacity", [
        [
            "bg-opacity",
            [
                "--tw-bg-opacity"
            ]
        ]
    ]),
    backgroundImage: (0, _createUtilityPlugin.default)("backgroundImage", [
        [
            "bg",
            [
                "background-image"
            ]
        ]
    ], {
        type: [
            "lookup",
            "image",
            "url"
        ]
    ***REMOVED***),
    gradientColorStops: (()=>{
        function transparentTo(value) {
            return (0, _withAlphaVariable.withAlphaValue)(value, 0, "rgb(255 255 255 / 0)");
        ***REMOVED***
        return function({ matchUtilities , theme , addDefaults  ***REMOVED***) {
            addDefaults("gradient-color-stops", {
                "--tw-gradient-from-position": " ",
                "--tw-gradient-via-position": " ",
                "--tw-gradient-to-position": " "
            ***REMOVED***);
            let options = {
                values: (0, _flattenColorPalette.default)(theme("gradientColorStops")),
                type: [
                    "color",
                    "any"
                ]
            ***REMOVED***;
            let positionOptions = {
                values: theme("gradientColorStopPositions"),
                type: [
                    "length",
                    "percentage"
                ]
            ***REMOVED***;
            matchUtilities({
                from: (value)=>{
                    let transparentToValue = transparentTo(value);
                    return {
                        "@defaults gradient-color-stops": {***REMOVED***,
                        "--tw-gradient-from": `${(0, _toColorValue.default)(value)***REMOVED*** var(--tw-gradient-from-position)`,
                        "--tw-gradient-to": `${transparentToValue***REMOVED*** var(--tw-gradient-to-position)`,
                        "--tw-gradient-stops": `var(--tw-gradient-from), var(--tw-gradient-to)`
                    ***REMOVED***;
                ***REMOVED***
            ***REMOVED***, options);
            matchUtilities({
                from: (value)=>{
                    return {
                        "--tw-gradient-from-position": value
                    ***REMOVED***;
                ***REMOVED***
            ***REMOVED***, positionOptions);
            matchUtilities({
                via: (value)=>{
                    let transparentToValue = transparentTo(value);
                    return {
                        "@defaults gradient-color-stops": {***REMOVED***,
                        "--tw-gradient-to": `${transparentToValue***REMOVED***  var(--tw-gradient-to-position)`,
                        "--tw-gradient-stops": `var(--tw-gradient-from), ${(0, _toColorValue.default)(value)***REMOVED*** var(--tw-gradient-via-position), var(--tw-gradient-to)`
                    ***REMOVED***;
                ***REMOVED***
            ***REMOVED***, options);
            matchUtilities({
                via: (value)=>{
                    return {
                        "--tw-gradient-via-position": value
                    ***REMOVED***;
                ***REMOVED***
            ***REMOVED***, positionOptions);
            matchUtilities({
                to: (value)=>({
                        "@defaults gradient-color-stops": {***REMOVED***,
                        "--tw-gradient-to": `${(0, _toColorValue.default)(value)***REMOVED*** var(--tw-gradient-to-position)`
                    ***REMOVED***)
            ***REMOVED***, options);
            matchUtilities({
                to: (value)=>{
                    return {
                        "--tw-gradient-to-position": value
                    ***REMOVED***;
                ***REMOVED***
            ***REMOVED***, positionOptions);
        ***REMOVED***;
    ***REMOVED***)(),
    boxDecorationBreak: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".decoration-slice": {
                "box-decoration-break": "slice"
            ***REMOVED***,
            ".decoration-clone": {
                "box-decoration-break": "clone"
            ***REMOVED***,
            ".box-decoration-slice": {
                "box-decoration-break": "slice"
            ***REMOVED***,
            ".box-decoration-clone": {
                "box-decoration-break": "clone"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    backgroundSize: (0, _createUtilityPlugin.default)("backgroundSize", [
        [
            "bg",
            [
                "background-size"
            ]
        ]
    ], {
        type: [
            "lookup",
            "length",
            "percentage",
            "size"
        ]
    ***REMOVED***),
    backgroundAttachment: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".bg-fixed": {
                "background-attachment": "fixed"
            ***REMOVED***,
            ".bg-local": {
                "background-attachment": "local"
            ***REMOVED***,
            ".bg-scroll": {
                "background-attachment": "scroll"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    backgroundClip: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".bg-clip-border": {
                "background-clip": "border-box"
            ***REMOVED***,
            ".bg-clip-padding": {
                "background-clip": "padding-box"
            ***REMOVED***,
            ".bg-clip-content": {
                "background-clip": "content-box"
            ***REMOVED***,
            ".bg-clip-text": {
                "background-clip": "text"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    backgroundPosition: (0, _createUtilityPlugin.default)("backgroundPosition", [
        [
            "bg",
            [
                "background-position"
            ]
        ]
    ], {
        type: [
            "lookup",
            [
                "position",
                {
                    preferOnConflict: true
                ***REMOVED***
            ]
        ]
    ***REMOVED***),
    backgroundRepeat: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".bg-repeat": {
                "background-repeat": "repeat"
            ***REMOVED***,
            ".bg-no-repeat": {
                "background-repeat": "no-repeat"
            ***REMOVED***,
            ".bg-repeat-x": {
                "background-repeat": "repeat-x"
            ***REMOVED***,
            ".bg-repeat-y": {
                "background-repeat": "repeat-y"
            ***REMOVED***,
            ".bg-repeat-round": {
                "background-repeat": "round"
            ***REMOVED***,
            ".bg-repeat-space": {
                "background-repeat": "space"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    backgroundOrigin: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".bg-origin-border": {
                "background-origin": "border-box"
            ***REMOVED***,
            ".bg-origin-padding": {
                "background-origin": "padding-box"
            ***REMOVED***,
            ".bg-origin-content": {
                "background-origin": "content-box"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    fill: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            fill: (value)=>{
                return {
                    fill: (0, _toColorValue.default)(value)
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: (0, _flattenColorPalette.default)(theme("fill")),
            type: [
                "color",
                "any"
            ]
        ***REMOVED***);
    ***REMOVED***,
    stroke: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            stroke: (value)=>{
                return {
                    stroke: (0, _toColorValue.default)(value)
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: (0, _flattenColorPalette.default)(theme("stroke")),
            type: [
                "color",
                "url",
                "any"
            ]
        ***REMOVED***);
    ***REMOVED***,
    strokeWidth: (0, _createUtilityPlugin.default)("strokeWidth", [
        [
            "stroke",
            [
                "stroke-width"
            ]
        ]
    ], {
        type: [
            "length",
            "number",
            "percentage"
        ]
    ***REMOVED***),
    objectFit: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".object-contain": {
                "object-fit": "contain"
            ***REMOVED***,
            ".object-cover": {
                "object-fit": "cover"
            ***REMOVED***,
            ".object-fill": {
                "object-fit": "fill"
            ***REMOVED***,
            ".object-none": {
                "object-fit": "none"
            ***REMOVED***,
            ".object-scale-down": {
                "object-fit": "scale-down"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    objectPosition: (0, _createUtilityPlugin.default)("objectPosition", [
        [
            "object",
            [
                "object-position"
            ]
        ]
    ]),
    padding: (0, _createUtilityPlugin.default)("padding", [
        [
            "p",
            [
                "padding"
            ]
        ],
        [
            [
                "px",
                [
                    "padding-left",
                    "padding-right"
                ]
            ],
            [
                "py",
                [
                    "padding-top",
                    "padding-bottom"
                ]
            ]
        ],
        [
            [
                "ps",
                [
                    "padding-inline-start"
                ]
            ],
            [
                "pe",
                [
                    "padding-inline-end"
                ]
            ],
            [
                "pt",
                [
                    "padding-top"
                ]
            ],
            [
                "pr",
                [
                    "padding-right"
                ]
            ],
            [
                "pb",
                [
                    "padding-bottom"
                ]
            ],
            [
                "pl",
                [
                    "padding-left"
                ]
            ]
        ]
    ]),
    textAlign: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".text-left": {
                "text-align": "left"
            ***REMOVED***,
            ".text-center": {
                "text-align": "center"
            ***REMOVED***,
            ".text-right": {
                "text-align": "right"
            ***REMOVED***,
            ".text-justify": {
                "text-align": "justify"
            ***REMOVED***,
            ".text-start": {
                "text-align": "start"
            ***REMOVED***,
            ".text-end": {
                "text-align": "end"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    textIndent: (0, _createUtilityPlugin.default)("textIndent", [
        [
            "indent",
            [
                "text-indent"
            ]
        ]
    ], {
        supportsNegativeValues: true
    ***REMOVED***),
    verticalAlign: ({ addUtilities , matchUtilities  ***REMOVED***)=>{
        addUtilities({
            ".align-baseline": {
                "vertical-align": "baseline"
            ***REMOVED***,
            ".align-top": {
                "vertical-align": "top"
            ***REMOVED***,
            ".align-middle": {
                "vertical-align": "middle"
            ***REMOVED***,
            ".align-bottom": {
                "vertical-align": "bottom"
            ***REMOVED***,
            ".align-text-top": {
                "vertical-align": "text-top"
            ***REMOVED***,
            ".align-text-bottom": {
                "vertical-align": "text-bottom"
            ***REMOVED***,
            ".align-sub": {
                "vertical-align": "sub"
            ***REMOVED***,
            ".align-super": {
                "vertical-align": "super"
            ***REMOVED***
        ***REMOVED***);
        matchUtilities({
            align: (value)=>({
                    "vertical-align": value
                ***REMOVED***)
        ***REMOVED***);
    ***REMOVED***,
    fontFamily: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            font: (value)=>{
                let [families, options = {***REMOVED***] = Array.isArray(value) && (0, _isPlainObject.default)(value[1]) ? value : [
                    value
                ];
                let { fontFeatureSettings , fontVariationSettings  ***REMOVED*** = options;
                return {
                    "font-family": Array.isArray(families) ? families.join(", ") : families,
                    ...fontFeatureSettings === undefined ? {***REMOVED*** : {
                        "font-feature-settings": fontFeatureSettings
                    ***REMOVED***,
                    ...fontVariationSettings === undefined ? {***REMOVED*** : {
                        "font-variation-settings": fontVariationSettings
                    ***REMOVED***
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("fontFamily"),
            type: [
                "lookup",
                "generic-name",
                "family-name"
            ]
        ***REMOVED***);
    ***REMOVED***,
    fontSize: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            text: (value, { modifier  ***REMOVED***)=>{
                let [fontSize, options] = Array.isArray(value) ? value : [
                    value
                ];
                if (modifier) {
                    return {
                        "font-size": fontSize,
                        "line-height": modifier
                    ***REMOVED***;
                ***REMOVED***
                let { lineHeight , letterSpacing , fontWeight  ***REMOVED*** = (0, _isPlainObject.default)(options) ? options : {
                    lineHeight: options
                ***REMOVED***;
                return {
                    "font-size": fontSize,
                    ...lineHeight === undefined ? {***REMOVED*** : {
                        "line-height": lineHeight
                    ***REMOVED***,
                    ...letterSpacing === undefined ? {***REMOVED*** : {
                        "letter-spacing": letterSpacing
                    ***REMOVED***,
                    ...fontWeight === undefined ? {***REMOVED*** : {
                        "font-weight": fontWeight
                    ***REMOVED***
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("fontSize"),
            modifiers: theme("lineHeight"),
            type: [
                "absolute-size",
                "relative-size",
                "length",
                "percentage"
            ]
        ***REMOVED***);
    ***REMOVED***,
    fontWeight: (0, _createUtilityPlugin.default)("fontWeight", [
        [
            "font",
            [
                "fontWeight"
            ]
        ]
    ], {
        type: [
            "lookup",
            "number",
            "any"
        ]
    ***REMOVED***),
    textTransform: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".uppercase": {
                "text-transform": "uppercase"
            ***REMOVED***,
            ".lowercase": {
                "text-transform": "lowercase"
            ***REMOVED***,
            ".capitalize": {
                "text-transform": "capitalize"
            ***REMOVED***,
            ".normal-case": {
                "text-transform": "none"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    fontStyle: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".italic": {
                "font-style": "italic"
            ***REMOVED***,
            ".not-italic": {
                "font-style": "normal"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    fontVariantNumeric: ({ addDefaults , addUtilities  ***REMOVED***)=>{
        let cssFontVariantNumericValue = "var(--tw-ordinal) var(--tw-slashed-zero) var(--tw-numeric-figure) var(--tw-numeric-spacing) var(--tw-numeric-fraction)";
        addDefaults("font-variant-numeric", {
            "--tw-ordinal": " ",
            "--tw-slashed-zero": " ",
            "--tw-numeric-figure": " ",
            "--tw-numeric-spacing": " ",
            "--tw-numeric-fraction": " "
        ***REMOVED***);
        addUtilities({
            ".normal-nums": {
                "font-variant-numeric": "normal"
            ***REMOVED***,
            ".ordinal": {
                "@defaults font-variant-numeric": {***REMOVED***,
                "--tw-ordinal": "ordinal",
                "font-variant-numeric": cssFontVariantNumericValue
            ***REMOVED***,
            ".slashed-zero": {
                "@defaults font-variant-numeric": {***REMOVED***,
                "--tw-slashed-zero": "slashed-zero",
                "font-variant-numeric": cssFontVariantNumericValue
            ***REMOVED***,
            ".lining-nums": {
                "@defaults font-variant-numeric": {***REMOVED***,
                "--tw-numeric-figure": "lining-nums",
                "font-variant-numeric": cssFontVariantNumericValue
            ***REMOVED***,
            ".oldstyle-nums": {
                "@defaults font-variant-numeric": {***REMOVED***,
                "--tw-numeric-figure": "oldstyle-nums",
                "font-variant-numeric": cssFontVariantNumericValue
            ***REMOVED***,
            ".proportional-nums": {
                "@defaults font-variant-numeric": {***REMOVED***,
                "--tw-numeric-spacing": "proportional-nums",
                "font-variant-numeric": cssFontVariantNumericValue
            ***REMOVED***,
            ".tabular-nums": {
                "@defaults font-variant-numeric": {***REMOVED***,
                "--tw-numeric-spacing": "tabular-nums",
                "font-variant-numeric": cssFontVariantNumericValue
            ***REMOVED***,
            ".diagonal-fractions": {
                "@defaults font-variant-numeric": {***REMOVED***,
                "--tw-numeric-fraction": "diagonal-fractions",
                "font-variant-numeric": cssFontVariantNumericValue
            ***REMOVED***,
            ".stacked-fractions": {
                "@defaults font-variant-numeric": {***REMOVED***,
                "--tw-numeric-fraction": "stacked-fractions",
                "font-variant-numeric": cssFontVariantNumericValue
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    lineHeight: (0, _createUtilityPlugin.default)("lineHeight", [
        [
            "leading",
            [
                "lineHeight"
            ]
        ]
    ]),
    letterSpacing: (0, _createUtilityPlugin.default)("letterSpacing", [
        [
            "tracking",
            [
                "letterSpacing"
            ]
        ]
    ], {
        supportsNegativeValues: true
    ***REMOVED***),
    textColor: ({ matchUtilities , theme , corePlugins  ***REMOVED***)=>{
        matchUtilities({
            text: (value)=>{
                if (!corePlugins("textOpacity")) {
                    return {
                        color: (0, _toColorValue.default)(value)
                    ***REMOVED***;
                ***REMOVED***
                return (0, _withAlphaVariable.default)({
                    color: value,
                    property: "color",
                    variable: "--tw-text-opacity"
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***, {
            values: (0, _flattenColorPalette.default)(theme("textColor")),
            type: [
                "color",
                "any"
            ]
        ***REMOVED***);
    ***REMOVED***,
    textOpacity: (0, _createUtilityPlugin.default)("textOpacity", [
        [
            "text-opacity",
            [
                "--tw-text-opacity"
            ]
        ]
    ]),
    textDecoration: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".underline": {
                "text-decoration-line": "underline"
            ***REMOVED***,
            ".overline": {
                "text-decoration-line": "overline"
            ***REMOVED***,
            ".line-through": {
                "text-decoration-line": "line-through"
            ***REMOVED***,
            ".no-underline": {
                "text-decoration-line": "none"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    textDecorationColor: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            decoration: (value)=>{
                return {
                    "text-decoration-color": (0, _toColorValue.default)(value)
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: (0, _flattenColorPalette.default)(theme("textDecorationColor")),
            type: [
                "color",
                "any"
            ]
        ***REMOVED***);
    ***REMOVED***,
    textDecorationStyle: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".decoration-solid": {
                "text-decoration-style": "solid"
            ***REMOVED***,
            ".decoration-double": {
                "text-decoration-style": "double"
            ***REMOVED***,
            ".decoration-dotted": {
                "text-decoration-style": "dotted"
            ***REMOVED***,
            ".decoration-dashed": {
                "text-decoration-style": "dashed"
            ***REMOVED***,
            ".decoration-wavy": {
                "text-decoration-style": "wavy"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    textDecorationThickness: (0, _createUtilityPlugin.default)("textDecorationThickness", [
        [
            "decoration",
            [
                "text-decoration-thickness"
            ]
        ]
    ], {
        type: [
            "length",
            "percentage"
        ]
    ***REMOVED***),
    textUnderlineOffset: (0, _createUtilityPlugin.default)("textUnderlineOffset", [
        [
            "underline-offset",
            [
                "text-underline-offset"
            ]
        ]
    ], {
        type: [
            "length",
            "percentage",
            "any"
        ]
    ***REMOVED***),
    fontSmoothing: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".antialiased": {
                "-webkit-font-smoothing": "antialiased",
                "-moz-osx-font-smoothing": "grayscale"
            ***REMOVED***,
            ".subpixel-antialiased": {
                "-webkit-font-smoothing": "auto",
                "-moz-osx-font-smoothing": "auto"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    placeholderColor: ({ matchUtilities , theme , corePlugins  ***REMOVED***)=>{
        matchUtilities({
            placeholder: (value)=>{
                if (!corePlugins("placeholderOpacity")) {
                    return {
                        "&::placeholder": {
                            color: (0, _toColorValue.default)(value)
                        ***REMOVED***
                    ***REMOVED***;
                ***REMOVED***
                return {
                    "&::placeholder": (0, _withAlphaVariable.default)({
                        color: value,
                        property: "color",
                        variable: "--tw-placeholder-opacity"
                    ***REMOVED***)
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: (0, _flattenColorPalette.default)(theme("placeholderColor")),
            type: [
                "color",
                "any"
            ]
        ***REMOVED***);
    ***REMOVED***,
    placeholderOpacity: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            "placeholder-opacity": (value)=>{
                return {
                    ["&::placeholder"]: {
                        "--tw-placeholder-opacity": value
                    ***REMOVED***
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("placeholderOpacity")
        ***REMOVED***);
    ***REMOVED***,
    caretColor: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            caret: (value)=>{
                return {
                    "caret-color": (0, _toColorValue.default)(value)
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: (0, _flattenColorPalette.default)(theme("caretColor")),
            type: [
                "color",
                "any"
            ]
        ***REMOVED***);
    ***REMOVED***,
    accentColor: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            accent: (value)=>{
                return {
                    "accent-color": (0, _toColorValue.default)(value)
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: (0, _flattenColorPalette.default)(theme("accentColor")),
            type: [
                "color",
                "any"
            ]
        ***REMOVED***);
    ***REMOVED***,
    opacity: (0, _createUtilityPlugin.default)("opacity", [
        [
            "opacity",
            [
                "opacity"
            ]
        ]
    ]),
    backgroundBlendMode: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".bg-blend-normal": {
                "background-blend-mode": "normal"
            ***REMOVED***,
            ".bg-blend-multiply": {
                "background-blend-mode": "multiply"
            ***REMOVED***,
            ".bg-blend-screen": {
                "background-blend-mode": "screen"
            ***REMOVED***,
            ".bg-blend-overlay": {
                "background-blend-mode": "overlay"
            ***REMOVED***,
            ".bg-blend-darken": {
                "background-blend-mode": "darken"
            ***REMOVED***,
            ".bg-blend-lighten": {
                "background-blend-mode": "lighten"
            ***REMOVED***,
            ".bg-blend-color-dodge": {
                "background-blend-mode": "color-dodge"
            ***REMOVED***,
            ".bg-blend-color-burn": {
                "background-blend-mode": "color-burn"
            ***REMOVED***,
            ".bg-blend-hard-light": {
                "background-blend-mode": "hard-light"
            ***REMOVED***,
            ".bg-blend-soft-light": {
                "background-blend-mode": "soft-light"
            ***REMOVED***,
            ".bg-blend-difference": {
                "background-blend-mode": "difference"
            ***REMOVED***,
            ".bg-blend-exclusion": {
                "background-blend-mode": "exclusion"
            ***REMOVED***,
            ".bg-blend-hue": {
                "background-blend-mode": "hue"
            ***REMOVED***,
            ".bg-blend-saturation": {
                "background-blend-mode": "saturation"
            ***REMOVED***,
            ".bg-blend-color": {
                "background-blend-mode": "color"
            ***REMOVED***,
            ".bg-blend-luminosity": {
                "background-blend-mode": "luminosity"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    mixBlendMode: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".mix-blend-normal": {
                "mix-blend-mode": "normal"
            ***REMOVED***,
            ".mix-blend-multiply": {
                "mix-blend-mode": "multiply"
            ***REMOVED***,
            ".mix-blend-screen": {
                "mix-blend-mode": "screen"
            ***REMOVED***,
            ".mix-blend-overlay": {
                "mix-blend-mode": "overlay"
            ***REMOVED***,
            ".mix-blend-darken": {
                "mix-blend-mode": "darken"
            ***REMOVED***,
            ".mix-blend-lighten": {
                "mix-blend-mode": "lighten"
            ***REMOVED***,
            ".mix-blend-color-dodge": {
                "mix-blend-mode": "color-dodge"
            ***REMOVED***,
            ".mix-blend-color-burn": {
                "mix-blend-mode": "color-burn"
            ***REMOVED***,
            ".mix-blend-hard-light": {
                "mix-blend-mode": "hard-light"
            ***REMOVED***,
            ".mix-blend-soft-light": {
                "mix-blend-mode": "soft-light"
            ***REMOVED***,
            ".mix-blend-difference": {
                "mix-blend-mode": "difference"
            ***REMOVED***,
            ".mix-blend-exclusion": {
                "mix-blend-mode": "exclusion"
            ***REMOVED***,
            ".mix-blend-hue": {
                "mix-blend-mode": "hue"
            ***REMOVED***,
            ".mix-blend-saturation": {
                "mix-blend-mode": "saturation"
            ***REMOVED***,
            ".mix-blend-color": {
                "mix-blend-mode": "color"
            ***REMOVED***,
            ".mix-blend-luminosity": {
                "mix-blend-mode": "luminosity"
            ***REMOVED***,
            ".mix-blend-plus-darker": {
                "mix-blend-mode": "plus-darker"
            ***REMOVED***,
            ".mix-blend-plus-lighter": {
                "mix-blend-mode": "plus-lighter"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    boxShadow: (()=>{
        let transformValue = (0, _transformThemeValue.default)("boxShadow");
        let defaultBoxShadow = [
            `var(--tw-ring-offset-shadow, 0 0 #0000)`,
            `var(--tw-ring-shadow, 0 0 #0000)`,
            `var(--tw-shadow)`
        ].join(", ");
        return function({ matchUtilities , addDefaults , theme  ***REMOVED***) {
            addDefaults("box-shadow", {
                "--tw-ring-offset-shadow": "0 0 #0000",
                "--tw-ring-shadow": "0 0 #0000",
                "--tw-shadow": "0 0 #0000",
                "--tw-shadow-colored": "0 0 #0000"
            ***REMOVED***);
            matchUtilities({
                shadow: (value)=>{
                    value = transformValue(value);
                    let ast = (0, _parseBoxShadowValue.parseBoxShadowValue)(value);
                    for (let shadow of ast){
                        // Don't override color if the whole shadow is a variable
                        if (!shadow.valid) {
                            continue;
                        ***REMOVED***
                        shadow.color = "var(--tw-shadow-color)";
                    ***REMOVED***
                    return {
                        "@defaults box-shadow": {***REMOVED***,
                        "--tw-shadow": value === "none" ? "0 0 #0000" : value,
                        "--tw-shadow-colored": value === "none" ? "0 0 #0000" : (0, _parseBoxShadowValue.formatBoxShadowValue)(ast),
                        "box-shadow": defaultBoxShadow
                    ***REMOVED***;
                ***REMOVED***
            ***REMOVED***, {
                values: theme("boxShadow"),
                type: [
                    "shadow"
                ]
            ***REMOVED***);
        ***REMOVED***;
    ***REMOVED***)(),
    boxShadowColor: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            shadow: (value)=>{
                return {
                    "--tw-shadow-color": (0, _toColorValue.default)(value),
                    "--tw-shadow": "var(--tw-shadow-colored)"
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: (0, _flattenColorPalette.default)(theme("boxShadowColor")),
            type: [
                "color",
                "any"
            ]
        ***REMOVED***);
    ***REMOVED***,
    outlineStyle: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".outline-none": {
                outline: "2px solid transparent",
                "outline-offset": "2px"
            ***REMOVED***,
            ".outline": {
                "outline-style": "solid"
            ***REMOVED***,
            ".outline-dashed": {
                "outline-style": "dashed"
            ***REMOVED***,
            ".outline-dotted": {
                "outline-style": "dotted"
            ***REMOVED***,
            ".outline-double": {
                "outline-style": "double"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    outlineWidth: (0, _createUtilityPlugin.default)("outlineWidth", [
        [
            "outline",
            [
                "outline-width"
            ]
        ]
    ], {
        type: [
            "length",
            "number",
            "percentage"
        ]
    ***REMOVED***),
    outlineOffset: (0, _createUtilityPlugin.default)("outlineOffset", [
        [
            "outline-offset",
            [
                "outline-offset"
            ]
        ]
    ], {
        type: [
            "length",
            "number",
            "percentage",
            "any"
        ],
        supportsNegativeValues: true
    ***REMOVED***),
    outlineColor: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            outline: (value)=>{
                return {
                    "outline-color": (0, _toColorValue.default)(value)
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: (0, _flattenColorPalette.default)(theme("outlineColor")),
            type: [
                "color",
                "any"
            ]
        ***REMOVED***);
    ***REMOVED***,
    ringWidth: ({ matchUtilities , addDefaults , addUtilities , theme , config  ***REMOVED***)=>{
        let ringColorDefault = (()=>{
            var _theme, _theme1;
            if ((0, _featureFlags.flagEnabled)(config(), "respectDefaultRingColorOpacity")) {
                return theme("ringColor.DEFAULT");
            ***REMOVED***
            let ringOpacityDefault = theme("ringOpacity.DEFAULT", "0.5");
            if (!((_theme = theme("ringColor")) === null || _theme === void 0 ? void 0 : _theme.DEFAULT)) {
                return `rgb(147 197 253 / ${ringOpacityDefault***REMOVED***)`;
            ***REMOVED***
            return (0, _withAlphaVariable.withAlphaValue)((_theme1 = theme("ringColor")) === null || _theme1 === void 0 ? void 0 : _theme1.DEFAULT, ringOpacityDefault, `rgb(147 197 253 / ${ringOpacityDefault***REMOVED***)`);
        ***REMOVED***)();
        addDefaults("ring-width", {
            "--tw-ring-inset": " ",
            "--tw-ring-offset-width": theme("ringOffsetWidth.DEFAULT", "0px"),
            "--tw-ring-offset-color": theme("ringOffsetColor.DEFAULT", "#fff"),
            "--tw-ring-color": ringColorDefault,
            "--tw-ring-offset-shadow": "0 0 #0000",
            "--tw-ring-shadow": "0 0 #0000",
            "--tw-shadow": "0 0 #0000",
            "--tw-shadow-colored": "0 0 #0000"
        ***REMOVED***);
        matchUtilities({
            ring: (value)=>{
                return {
                    "@defaults ring-width": {***REMOVED***,
                    "--tw-ring-offset-shadow": `var(--tw-ring-inset) 0 0 0 var(--tw-ring-offset-width) var(--tw-ring-offset-color)`,
                    "--tw-ring-shadow": `var(--tw-ring-inset) 0 0 0 calc(${value***REMOVED*** + var(--tw-ring-offset-width)) var(--tw-ring-color)`,
                    "box-shadow": [
                        `var(--tw-ring-offset-shadow)`,
                        `var(--tw-ring-shadow)`,
                        `var(--tw-shadow, 0 0 #0000)`
                    ].join(", ")
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("ringWidth"),
            type: "length"
        ***REMOVED***);
        addUtilities({
            ".ring-inset": {
                "@defaults ring-width": {***REMOVED***,
                "--tw-ring-inset": "inset"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    ringColor: ({ matchUtilities , theme , corePlugins  ***REMOVED***)=>{
        matchUtilities({
            ring: (value)=>{
                if (!corePlugins("ringOpacity")) {
                    return {
                        "--tw-ring-color": (0, _toColorValue.default)(value)
                    ***REMOVED***;
                ***REMOVED***
                return (0, _withAlphaVariable.default)({
                    color: value,
                    property: "--tw-ring-color",
                    variable: "--tw-ring-opacity"
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***, {
            values: Object.fromEntries(Object.entries((0, _flattenColorPalette.default)(theme("ringColor"))).filter(([modifier])=>modifier !== "DEFAULT")),
            type: [
                "color",
                "any"
            ]
        ***REMOVED***);
    ***REMOVED***,
    ringOpacity: (helpers)=>{
        let { config  ***REMOVED*** = helpers;
        return (0, _createUtilityPlugin.default)("ringOpacity", [
            [
                "ring-opacity",
                [
                    "--tw-ring-opacity"
                ]
            ]
        ], {
            filterDefault: !(0, _featureFlags.flagEnabled)(config(), "respectDefaultRingColorOpacity")
        ***REMOVED***)(helpers);
    ***REMOVED***,
    ringOffsetWidth: (0, _createUtilityPlugin.default)("ringOffsetWidth", [
        [
            "ring-offset",
            [
                "--tw-ring-offset-width"
            ]
        ]
    ], {
        type: "length"
    ***REMOVED***),
    ringOffsetColor: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            "ring-offset": (value)=>{
                return {
                    "--tw-ring-offset-color": (0, _toColorValue.default)(value)
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: (0, _flattenColorPalette.default)(theme("ringOffsetColor")),
            type: [
                "color",
                "any"
            ]
        ***REMOVED***);
    ***REMOVED***,
    blur: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            blur: (value)=>{
                return {
                    "--tw-blur": value.trim() === "" ? " " : `blur(${value***REMOVED***)`,
                    "@defaults filter": {***REMOVED***,
                    filter: cssFilterValue
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("blur")
        ***REMOVED***);
    ***REMOVED***,
    brightness: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            brightness: (value)=>{
                return {
                    "--tw-brightness": `brightness(${value***REMOVED***)`,
                    "@defaults filter": {***REMOVED***,
                    filter: cssFilterValue
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("brightness")
        ***REMOVED***);
    ***REMOVED***,
    contrast: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            contrast: (value)=>{
                return {
                    "--tw-contrast": `contrast(${value***REMOVED***)`,
                    "@defaults filter": {***REMOVED***,
                    filter: cssFilterValue
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("contrast")
        ***REMOVED***);
    ***REMOVED***,
    dropShadow: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            "drop-shadow": (value)=>{
                return {
                    "--tw-drop-shadow": Array.isArray(value) ? value.map((v)=>`drop-shadow(${v***REMOVED***)`).join(" ") : `drop-shadow(${value***REMOVED***)`,
                    "@defaults filter": {***REMOVED***,
                    filter: cssFilterValue
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("dropShadow")
        ***REMOVED***);
    ***REMOVED***,
    grayscale: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            grayscale: (value)=>{
                return {
                    "--tw-grayscale": `grayscale(${value***REMOVED***)`,
                    "@defaults filter": {***REMOVED***,
                    filter: cssFilterValue
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("grayscale")
        ***REMOVED***);
    ***REMOVED***,
    hueRotate: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            "hue-rotate": (value)=>{
                return {
                    "--tw-hue-rotate": `hue-rotate(${value***REMOVED***)`,
                    "@defaults filter": {***REMOVED***,
                    filter: cssFilterValue
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("hueRotate"),
            supportsNegativeValues: true
        ***REMOVED***);
    ***REMOVED***,
    invert: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            invert: (value)=>{
                return {
                    "--tw-invert": `invert(${value***REMOVED***)`,
                    "@defaults filter": {***REMOVED***,
                    filter: cssFilterValue
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("invert")
        ***REMOVED***);
    ***REMOVED***,
    saturate: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            saturate: (value)=>{
                return {
                    "--tw-saturate": `saturate(${value***REMOVED***)`,
                    "@defaults filter": {***REMOVED***,
                    filter: cssFilterValue
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("saturate")
        ***REMOVED***);
    ***REMOVED***,
    sepia: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            sepia: (value)=>{
                return {
                    "--tw-sepia": `sepia(${value***REMOVED***)`,
                    "@defaults filter": {***REMOVED***,
                    filter: cssFilterValue
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("sepia")
        ***REMOVED***);
    ***REMOVED***,
    filter: ({ addDefaults , addUtilities  ***REMOVED***)=>{
        addDefaults("filter", {
            "--tw-blur": " ",
            "--tw-brightness": " ",
            "--tw-contrast": " ",
            "--tw-grayscale": " ",
            "--tw-hue-rotate": " ",
            "--tw-invert": " ",
            "--tw-saturate": " ",
            "--tw-sepia": " ",
            "--tw-drop-shadow": " "
        ***REMOVED***);
        addUtilities({
            ".filter": {
                "@defaults filter": {***REMOVED***,
                filter: cssFilterValue
            ***REMOVED***,
            ".filter-none": {
                filter: "none"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    backdropBlur: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            "backdrop-blur": (value)=>{
                return {
                    "--tw-backdrop-blur": value.trim() === "" ? " " : `blur(${value***REMOVED***)`,
                    "@defaults backdrop-filter": {***REMOVED***,
                    "-webkit-backdrop-filter": cssBackdropFilterValue,
                    "backdrop-filter": cssBackdropFilterValue
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("backdropBlur")
        ***REMOVED***);
    ***REMOVED***,
    backdropBrightness: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            "backdrop-brightness": (value)=>{
                return {
                    "--tw-backdrop-brightness": `brightness(${value***REMOVED***)`,
                    "@defaults backdrop-filter": {***REMOVED***,
                    "-webkit-backdrop-filter": cssBackdropFilterValue,
                    "backdrop-filter": cssBackdropFilterValue
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("backdropBrightness")
        ***REMOVED***);
    ***REMOVED***,
    backdropContrast: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            "backdrop-contrast": (value)=>{
                return {
                    "--tw-backdrop-contrast": `contrast(${value***REMOVED***)`,
                    "@defaults backdrop-filter": {***REMOVED***,
                    "-webkit-backdrop-filter": cssBackdropFilterValue,
                    "backdrop-filter": cssBackdropFilterValue
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("backdropContrast")
        ***REMOVED***);
    ***REMOVED***,
    backdropGrayscale: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            "backdrop-grayscale": (value)=>{
                return {
                    "--tw-backdrop-grayscale": `grayscale(${value***REMOVED***)`,
                    "@defaults backdrop-filter": {***REMOVED***,
                    "-webkit-backdrop-filter": cssBackdropFilterValue,
                    "backdrop-filter": cssBackdropFilterValue
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("backdropGrayscale")
        ***REMOVED***);
    ***REMOVED***,
    backdropHueRotate: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            "backdrop-hue-rotate": (value)=>{
                return {
                    "--tw-backdrop-hue-rotate": `hue-rotate(${value***REMOVED***)`,
                    "@defaults backdrop-filter": {***REMOVED***,
                    "-webkit-backdrop-filter": cssBackdropFilterValue,
                    "backdrop-filter": cssBackdropFilterValue
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("backdropHueRotate"),
            supportsNegativeValues: true
        ***REMOVED***);
    ***REMOVED***,
    backdropInvert: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            "backdrop-invert": (value)=>{
                return {
                    "--tw-backdrop-invert": `invert(${value***REMOVED***)`,
                    "@defaults backdrop-filter": {***REMOVED***,
                    "-webkit-backdrop-filter": cssBackdropFilterValue,
                    "backdrop-filter": cssBackdropFilterValue
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("backdropInvert")
        ***REMOVED***);
    ***REMOVED***,
    backdropOpacity: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            "backdrop-opacity": (value)=>{
                return {
                    "--tw-backdrop-opacity": `opacity(${value***REMOVED***)`,
                    "@defaults backdrop-filter": {***REMOVED***,
                    "-webkit-backdrop-filter": cssBackdropFilterValue,
                    "backdrop-filter": cssBackdropFilterValue
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("backdropOpacity")
        ***REMOVED***);
    ***REMOVED***,
    backdropSaturate: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            "backdrop-saturate": (value)=>{
                return {
                    "--tw-backdrop-saturate": `saturate(${value***REMOVED***)`,
                    "@defaults backdrop-filter": {***REMOVED***,
                    "-webkit-backdrop-filter": cssBackdropFilterValue,
                    "backdrop-filter": cssBackdropFilterValue
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("backdropSaturate")
        ***REMOVED***);
    ***REMOVED***,
    backdropSepia: ({ matchUtilities , theme  ***REMOVED***)=>{
        matchUtilities({
            "backdrop-sepia": (value)=>{
                return {
                    "--tw-backdrop-sepia": `sepia(${value***REMOVED***)`,
                    "@defaults backdrop-filter": {***REMOVED***,
                    "-webkit-backdrop-filter": cssBackdropFilterValue,
                    "backdrop-filter": cssBackdropFilterValue
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("backdropSepia")
        ***REMOVED***);
    ***REMOVED***,
    backdropFilter: ({ addDefaults , addUtilities  ***REMOVED***)=>{
        addDefaults("backdrop-filter", {
            "--tw-backdrop-blur": " ",
            "--tw-backdrop-brightness": " ",
            "--tw-backdrop-contrast": " ",
            "--tw-backdrop-grayscale": " ",
            "--tw-backdrop-hue-rotate": " ",
            "--tw-backdrop-invert": " ",
            "--tw-backdrop-opacity": " ",
            "--tw-backdrop-saturate": " ",
            "--tw-backdrop-sepia": " "
        ***REMOVED***);
        addUtilities({
            ".backdrop-filter": {
                "@defaults backdrop-filter": {***REMOVED***,
                "-webkit-backdrop-filter": cssBackdropFilterValue,
                "backdrop-filter": cssBackdropFilterValue
            ***REMOVED***,
            ".backdrop-filter-none": {
                "-webkit-backdrop-filter": "none",
                "backdrop-filter": "none"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    transitionProperty: ({ matchUtilities , theme  ***REMOVED***)=>{
        let defaultTimingFunction = theme("transitionTimingFunction.DEFAULT");
        let defaultDuration = theme("transitionDuration.DEFAULT");
        matchUtilities({
            transition: (value)=>{
                return {
                    "transition-property": value,
                    ...value === "none" ? {***REMOVED*** : {
                        "transition-timing-function": defaultTimingFunction,
                        "transition-duration": defaultDuration
                    ***REMOVED***
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***, {
            values: theme("transitionProperty")
        ***REMOVED***);
    ***REMOVED***,
    transitionDelay: (0, _createUtilityPlugin.default)("transitionDelay", [
        [
            "delay",
            [
                "transitionDelay"
            ]
        ]
    ]),
    transitionDuration: (0, _createUtilityPlugin.default)("transitionDuration", [
        [
            "duration",
            [
                "transitionDuration"
            ]
        ]
    ], {
        filterDefault: true
    ***REMOVED***),
    transitionTimingFunction: (0, _createUtilityPlugin.default)("transitionTimingFunction", [
        [
            "ease",
            [
                "transitionTimingFunction"
            ]
        ]
    ], {
        filterDefault: true
    ***REMOVED***),
    willChange: (0, _createUtilityPlugin.default)("willChange", [
        [
            "will-change",
            [
                "will-change"
            ]
        ]
    ]),
    contain: ({ addDefaults , addUtilities  ***REMOVED***)=>{
        let cssContainValue = "var(--tw-contain-size) var(--tw-contain-layout) var(--tw-contain-paint) var(--tw-contain-style)";
        addDefaults("contain", {
            "--tw-contain-size": " ",
            "--tw-contain-layout": " ",
            "--tw-contain-paint": " ",
            "--tw-contain-style": " "
        ***REMOVED***);
        addUtilities({
            ".contain-none": {
                contain: "none"
            ***REMOVED***,
            ".contain-content": {
                contain: "content"
            ***REMOVED***,
            ".contain-strict": {
                contain: "strict"
            ***REMOVED***,
            ".contain-size": {
                "@defaults contain": {***REMOVED***,
                "--tw-contain-size": "size",
                contain: cssContainValue
            ***REMOVED***,
            ".contain-inline-size": {
                "@defaults contain": {***REMOVED***,
                "--tw-contain-size": "inline-size",
                contain: cssContainValue
            ***REMOVED***,
            ".contain-layout": {
                "@defaults contain": {***REMOVED***,
                "--tw-contain-layout": "layout",
                contain: cssContainValue
            ***REMOVED***,
            ".contain-paint": {
                "@defaults contain": {***REMOVED***,
                "--tw-contain-paint": "paint",
                contain: cssContainValue
            ***REMOVED***,
            ".contain-style": {
                "@defaults contain": {***REMOVED***,
                "--tw-contain-style": "style",
                contain: cssContainValue
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***,
    content: (0, _createUtilityPlugin.default)("content", [
        [
            "content",
            [
                "--tw-content",
                [
                    "content",
                    "var(--tw-content)"
                ]
            ]
        ]
    ]),
    forcedColorAdjust: ({ addUtilities  ***REMOVED***)=>{
        addUtilities({
            ".forced-color-adjust-auto": {
                "forced-color-adjust": "auto"
            ***REMOVED***,
            ".forced-color-adjust-none": {
                "forced-color-adjust": "none"
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***
***REMOVED***;
