// postcss-minify-selectors/src/index.js
'use strict';
const parser = require('postcss-selector-parser');
const canUnquote = require('./lib/canUnquote.js');

const pseudoElements = new Set([
  '::before',
  '::after',
  '::first-letter',
  '::first-line',
]);

/**
 * @param {parser.Attribute***REMOVED*** selector
 * @return {void***REMOVED***
 */
function attribute(selector) {
  if (selector.value) {
    if (selector.raws.value) {
      // Join selectors that are split over new lines
      selector.raws.value = selector.raws.value.replace(/\\\n/g, '').trim();
    ***REMOVED***
    if (canUnquote(selector.value)) {
      selector.quoteMark = null;
    ***REMOVED***

    if (selector.operator) {
      selector.operator = /** @type {parser.AttributeOperator***REMOVED*** */ (
        selector.operator.trim()
      );
    ***REMOVED***
  ***REMOVED***

  selector.rawSpaceBefore = '';
  selector.rawSpaceAfter = '';
  selector.spaces.attribute = { before: '', after: '' ***REMOVED***;
  selector.spaces.operator = { before: '', after: '' ***REMOVED***;
  selector.spaces.value = {
    before: '',
    after: selector.insensitive ? ' ' : '',
  ***REMOVED***;

  if (selector.raws.spaces) {
    selector.raws.spaces.attribute = {
      before: '',
      after: '',
    ***REMOVED***;

    selector.raws.spaces.operator = {
      before: '',
      after: '',
    ***REMOVED***;

    selector.raws.spaces.value = {
      before: '',
      after: selector.insensitive ? ' ' : '',
    ***REMOVED***;

    if (selector.insensitive) {
      selector.raws.spaces.insensitive = {
        before: '',
        after: '',
      ***REMOVED***;
    ***REMOVED***
  ***REMOVED***

  selector.attribute = selector.attribute.trim();
***REMOVED***

/**
 * @param {parser.Combinator***REMOVED*** selector
 * @return {void***REMOVED***
 */
function combinator(selector) {
  const value = selector.value.trim();
  selector.spaces.before = '';
  selector.spaces.after = '';
  selector.rawSpaceBefore = '';
  selector.rawSpaceAfter = '';
  selector.value = value.length ? value : ' ';
***REMOVED***

const pseudoReplacements = new Map([
  [':nth-child', ':first-child'],
  [':nth-of-type', ':first-of-type'],
  [':nth-last-child', ':last-child'],
  [':nth-last-of-type', ':last-of-type'],
]);

/**
 * @param {parser.Pseudo***REMOVED*** selector
 * @return {void***REMOVED***
 */
function pseudo(selector) {
  const value = selector.value.toLowerCase();

  if (selector.nodes.length === 1 && pseudoReplacements.has(value)) {
    const first = selector.at(0);
    const one = first.at(0);

    if (first.length === 1) {
      if (one.value === '1') {
        selector.replaceWith(
          parser.pseudo({
            value: /** @type {string***REMOVED*** */ (pseudoReplacements.get(value)),
          ***REMOVED***)
        );
      ***REMOVED***

      if (one.value && one.value.toLowerCase() === 'even') {
        one.value = '2n';
      ***REMOVED***
    ***REMOVED***

    if (first.length === 3) {
      const two = first.at(1);
      const three = first.at(2);

      if (
        one.value &&
        one.value.toLowerCase() === '2n' &&
        two.value === '+' &&
        three.value === '1'
      ) {
        one.value = 'odd';

        two.remove();
        three.remove();
      ***REMOVED***
    ***REMOVED***

    return;
  ***REMOVED***

  selector.walk((child) => {
    if (child.type === 'selector' && child.parent) {
      const uniques = new Set();
      child.parent.each((sibling) => {
        const siblingStr = String(sibling);

        if (!uniques.has(siblingStr)) {
          uniques.add(siblingStr);
        ***REMOVED*** else {
          sibling.remove();
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***);

  if (pseudoElements.has(value)) {
    selector.value = selector.value.slice(1);
  ***REMOVED***
***REMOVED***

const tagReplacements = new Map([
  ['from', '0%'],
  ['100%', 'to'],
]);

/**
 * @param {parser.Tag***REMOVED*** selector
 * @return {void***REMOVED***
 */
function tag(selector) {
  const value = selector.value.toLowerCase();

  if (tagReplacements.has(value)) {
    selector.value = /** @type {string***REMOVED*** */ (tagReplacements.get(value));
  ***REMOVED***
***REMOVED***

/**
 * @param {parser.Universal***REMOVED*** selector
 * @return {void***REMOVED***
 */
function universal(selector) {
  const next = selector.next();

  if (next && next.type !== 'combinator') {
    selector.remove();
  ***REMOVED***
***REMOVED***

const reducers = new Map(
  /** @type {[string, ((selector: parser.Node) => void)][]***REMOVED****/ ([
    ['attribute', attribute],
    ['combinator', combinator],
    ['pseudo', pseudo],
    ['tag', tag],
    ['universal', universal],
  ])
);

/**
 * @type {import('postcss').PluginCreator<void>***REMOVED***
 * @return {import('postcss').Plugin***REMOVED***
 */
function pluginCreator() {
  return {
    postcssPlugin: 'postcss-minify-selectors',

    OnceExit(css) {
      const cache = new Map();
      const processor = parser((selectors) => {
        const uniqueSelectors = new Set();

        selectors.walk((sel) => {
          // Trim whitespace around the value
          sel.spaces.before = sel.spaces.after = '';
          const reducer = reducers.get(sel.type);
          if (reducer !== undefined) {
            reducer(sel);
            return;
          ***REMOVED***

          const toString = String(sel);

          if (
            sel.type === 'selector' &&
            sel.parent &&
            sel.parent.type !== 'pseudo'
          ) {
            if (!uniqueSelectors.has(toString)) {
              uniqueSelectors.add(toString);
            ***REMOVED*** else {
              sel.remove();
            ***REMOVED***
          ***REMOVED***
        ***REMOVED***);
        selectors.nodes.sort();
      ***REMOVED***);

      css.walkRules((rule) => {
        const selector =
          rule.raws.selector && rule.raws.selector.value === rule.selector
            ? rule.raws.selector.raw
            : rule.selector;

        // If the selector ends with a ':' it is likely a part of a custom mixin,
        // so just pass through.
        if (selector[selector.length - 1] === ':') {
          return;
        ***REMOVED***

        if (cache.has(selector)) {
          rule.selector = cache.get(selector);

          return;
        ***REMOVED***

        const optimizedSelector = processor.processSync(selector);

        rule.selector = optimizedSelector;
        cache.set(selector, optimizedSelector);
      ***REMOVED***);
    ***REMOVED***,
  ***REMOVED***;
***REMOVED***

pluginCreator.postcss = true;
module.exports = pluginCreator;
