// jest-watch-typeahead/build/file_name_plugin/plugin.js
import { Prompt ***REMOVED*** from 'jest-watcher';
import FileNamePatternPrompt from "./prompt.js";
export default class FileNamePlugin {
  constructor({
    stdin,
    stdout,
    config = {***REMOVED***
  ***REMOVED***) {
    this._stdin = stdin;
    this._stdout = stdout;
    this._prompt = new Prompt();
    this._projects = [];
    this._usageInfo = {
      key: config.key || 'p',
      prompt: config.prompt || 'filter by a filename regex pattern'
    ***REMOVED***;
  ***REMOVED***

  apply(jestHooks) {
    jestHooks.onFileChange(({
      projects
    ***REMOVED***) => {
      this._projects = projects;
    ***REMOVED***);
  ***REMOVED***

  onKey(key) {
    this._prompt.put(key);
  ***REMOVED***

  run(globalConfig, updateConfigAndRun) {
    const p = new FileNamePatternPrompt(this._stdout, this._prompt);
    p.updateSearchSources(this._projects);
    return new Promise((res, rej) => {
      p.run(testPathPattern => {
        updateConfigAndRun({
          mode: 'watch',
          testPathPattern
        ***REMOVED***);
        res();
      ***REMOVED***, rej);
    ***REMOVED***);
  ***REMOVED***

  getUsageInfo() {
    return this._usageInfo;
  ***REMOVED***

***REMOVED***
