// jsdom/lib/jsdom/living/fetch/header-list.js
"use strict";

/**
 * Provides some utility functions for somewhat efficiently modifying a
 * collection of headers.
 *
 * Note that this class only operates on ByteStrings (which is also why we use
 * toLowerCase internally).
 */
class HeaderList {
  constructor() {
    this.headers = new Map();
  ***REMOVED***

  append(name, value) {
    const existing = this.headers.get(name.toLowerCase());
    if (existing) {
      name = existing[0].name;
      existing.push({ name, value ***REMOVED***);
    ***REMOVED*** else {
      this.headers.set(name.toLowerCase(), [{ name, value ***REMOVED***]);
    ***REMOVED***
  ***REMOVED***

  contains(name) {
    return this.headers.has(name.toLowerCase());
  ***REMOVED***

  get(name) {
    name = name.toLowerCase();
    const values = this.headers.get(name);
    if (!values) {
      return null;
    ***REMOVED***
    return values.map(h => h.value).join(", ");
  ***REMOVED***

  delete(name) {
    this.headers.delete(name.toLowerCase());
  ***REMOVED***

  set(name, value) {
    const lowerName = name.toLowerCase();
    this.headers.delete(lowerName);
    this.headers.set(lowerName, [{ name, value ***REMOVED***]);
  ***REMOVED***

  sortAndCombine() {
    const names = [...this.headers.keys()].sort();
    return names.map(n => [n, this.get(n)]);
  ***REMOVED***
***REMOVED***

module.exports = HeaderList;
