// sucrase/dist/esm/transformers/ReactHotLoaderTransformer.js
import {IdentifierRole, isTopLevelDeclaration***REMOVED*** from "../parser/tokenizer";

import Transformer from "./Transformer";

export default class ReactHotLoaderTransformer extends Transformer {
   __init() {this.extractedDefaultExportName = null***REMOVED***

  constructor( tokens,  filePath) {
    super();this.tokens = tokens;this.filePath = filePath;ReactHotLoaderTransformer.prototype.__init.call(this);;
  ***REMOVED***

  setExtractedDefaultExportName(extractedDefaultExportName) {
    this.extractedDefaultExportName = extractedDefaultExportName;
  ***REMOVED***

  getPrefixCode() {
    return `
      (function () {
        var enterModule = require('react-hot-loader').enterModule;
        enterModule && enterModule(module);
      ***REMOVED***)();`
      .replace(/\s+/g, " ")
      .trim();
  ***REMOVED***

  getSuffixCode() {
    const topLevelNames = new Set();
    for (const token of this.tokens.tokens) {
      if (
        !token.isType &&
        isTopLevelDeclaration(token) &&
        token.identifierRole !== IdentifierRole.ImportDeclaration
      ) {
        topLevelNames.add(this.tokens.identifierNameForToken(token));
      ***REMOVED***
    ***REMOVED***
    const namesToRegister = Array.from(topLevelNames).map((name) => ({
      variableName: name,
      uniqueLocalName: name,
    ***REMOVED***));
    if (this.extractedDefaultExportName) {
      namesToRegister.push({
        variableName: this.extractedDefaultExportName,
        uniqueLocalName: "default",
      ***REMOVED***);
    ***REMOVED***
    return `
;(function () {
  var reactHotLoader = require('react-hot-loader').default;
  var leaveModule = require('react-hot-loader').leaveModule;
  if (!reactHotLoader) {
    return;
  ***REMOVED***
${namesToRegister
  .map(
    ({variableName, uniqueLocalName***REMOVED***) =>
      `  reactHotLoader.register(${variableName***REMOVED***, "${uniqueLocalName***REMOVED***", ${JSON.stringify(
        this.filePath || "",
      )***REMOVED***);`,
  )
  .join("\n")***REMOVED***
  leaveModule(module);
***REMOVED***)();`;
  ***REMOVED***

  process() {
    return false;
  ***REMOVED***
***REMOVED***
