// ajv/dist/vocabularies/validation/enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
const codegen_1 = require("../../compile/codegen");
const util_1 = require("../../compile/util");
const equal_1 = require("../../runtime/equal");
const error = {
    message: "must be equal to one of the allowed values",
    params: ({ schemaCode ***REMOVED***) => (0, codegen_1._) `{allowedValues: ${schemaCode***REMOVED******REMOVED***`,
***REMOVED***;
const def = {
    keyword: "enum",
    schemaType: "array",
    $data: true,
    error,
    code(cxt) {
        const { gen, data, $data, schema, schemaCode, it ***REMOVED*** = cxt;
        if (!$data && schema.length === 0)
            throw new Error("enum must have non-empty array");
        const useLoop = schema.length >= it.opts.loopEnum;
        let eql;
        const getEql = () => (eql !== null && eql !== void 0 ? eql : (eql = (0, util_1.useFunc)(gen, equal_1.default)));
        let valid;
        if (useLoop || $data) {
            valid = gen.let("valid");
            cxt.block$data(valid, loopEnum);
        ***REMOVED***
        else {
            /* istanbul ignore if */
            if (!Array.isArray(schema))
                throw new Error("ajv implementation error");
            const vSchema = gen.const("vSchema", schemaCode);
            valid = (0, codegen_1.or)(...schema.map((_x, i) => equalCode(vSchema, i)));
        ***REMOVED***
        cxt.pass(valid);
        function loopEnum() {
            gen.assign(valid, false);
            gen.forOf("v", schemaCode, (v) => gen.if((0, codegen_1._) `${getEql()***REMOVED***(${data***REMOVED***, ${v***REMOVED***)`, () => gen.assign(valid, true).break()));
        ***REMOVED***
        function equalCode(vSchema, i) {
            const sch = schema[i];
            return typeof sch === "object" && sch !== null
                ? (0, codegen_1._) `${getEql()***REMOVED***(${data***REMOVED***, ${vSchema***REMOVED***[${i***REMOVED***])`
                : (0, codegen_1._) `${data***REMOVED*** === ${sch***REMOVED***`;
        ***REMOVED***
    ***REMOVED***,
***REMOVED***;
exports.default = def;
