// clean-css/lib/writer/one-time.js
var helpers = require('./helpers');

function store(serializeContext, token) {
  serializeContext.output.push(typeof token == 'string' ? token : token[1]);
***REMOVED***

function context() {
  var newContext = {
    output: [],
    store: store
  ***REMOVED***;

  return newContext;
***REMOVED***

function all(tokens) {
  var oneTimeContext = context();
  helpers.all(oneTimeContext, tokens);
  return oneTimeContext.output.join('');
***REMOVED***

function body(tokens) {
  var oneTimeContext = context();
  helpers.body(oneTimeContext, tokens);
  return oneTimeContext.output.join('');
***REMOVED***

function property(tokens, position) {
  var oneTimeContext = context();
  helpers.property(oneTimeContext, tokens, position, true);
  return oneTimeContext.output.join('');
***REMOVED***

function rules(tokens) {
  var oneTimeContext = context();
  helpers.rules(oneTimeContext, tokens);
  return oneTimeContext.output.join('');
***REMOVED***

function value(tokens) {
  var oneTimeContext = context();
  helpers.value(oneTimeContext, tokens);
  return oneTimeContext.output.join('');
***REMOVED***

module.exports = {
  all: all,
  body: body,
  property: property,
  rules: rules,
  value: value
***REMOVED***;
