// @webassemblyjs/wasm-parser/lib/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; ***REMOVED***; ***REMOVED*** else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; ***REMOVED***; ***REMOVED*** return _typeof(obj); ***REMOVED***

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.decode = decode;

var decoder = _interopRequireWildcard(require("./decoder"));

var t = _interopRequireWildcard(require("@webassemblyjs/ast"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; ***REMOVED***)(nodeInterop); ***REMOVED***

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; ***REMOVED*** if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj ***REMOVED***; ***REMOVED*** var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); ***REMOVED*** var newObj = {***REMOVED***; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); ***REMOVED*** else { newObj[key] = obj[key]; ***REMOVED*** ***REMOVED*** ***REMOVED*** newObj["default"] = obj; if (cache) { cache.set(obj, newObj); ***REMOVED*** return newObj; ***REMOVED***

/**
 * TODO(sven): I added initial props, but we should rather fix
 * https://github.com/xtuc/webassemblyjs/issues/405
 */
var defaultDecoderOpts = {
  dump: false,
  ignoreCodeSection: false,
  ignoreDataSection: false,
  ignoreCustomNameSection: false
***REMOVED***; // traverses the AST, locating function name metadata, which is then
// used to update index-based identifiers with function names

function restoreFunctionNames(ast) {
  var functionNames = [];
  t.traverse(ast, {
    FunctionNameMetadata: function FunctionNameMetadata(_ref) {
      var node = _ref.node;
      functionNames.push({
        name: node.value,
        index: node.index
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***);

  if (functionNames.length === 0) {
    return;
  ***REMOVED***

  t.traverse(ast, {
    Func: function (_Func) {
      function Func(_x) {
        return _Func.apply(this, arguments);
      ***REMOVED***

      Func.toString = function () {
        return _Func.toString();
      ***REMOVED***;

      return Func;
    ***REMOVED***(function (_ref2) {
      var node = _ref2.node;
      // $FlowIgnore
      var nodeName = node.name;
      var indexBasedFunctionName = nodeName.value;
      var index = Number(indexBasedFunctionName.replace("func_", ""));
      var functionName = functionNames.find(function (f) {
        return f.index === index;
      ***REMOVED***);

      if (functionName) {
        var oldValue = nodeName.value;
        nodeName.value = functionName.name; // $FlowIgnore

        nodeName.numeric = oldValue; // $FlowIgnore

        delete nodeName.raw;
      ***REMOVED***
    ***REMOVED***),
    // Also update the reference in the export
    ModuleExport: function (_ModuleExport) {
      function ModuleExport(_x2) {
        return _ModuleExport.apply(this, arguments);
      ***REMOVED***

      ModuleExport.toString = function () {
        return _ModuleExport.toString();
      ***REMOVED***;

      return ModuleExport;
    ***REMOVED***(function (_ref3) {
      var node = _ref3.node;

      if (node.descr.exportType === "Func") {
        // $FlowIgnore
        var nodeName = node.descr.id;
        var index = nodeName.value;
        var functionName = functionNames.find(function (f) {
          return f.index === index;
        ***REMOVED***);

        if (functionName) {
          node.descr.id = t.identifier(functionName.name);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***),
    ModuleImport: function (_ModuleImport) {
      function ModuleImport(_x3) {
        return _ModuleImport.apply(this, arguments);
      ***REMOVED***

      ModuleImport.toString = function () {
        return _ModuleImport.toString();
      ***REMOVED***;

      return ModuleImport;
    ***REMOVED***(function (_ref4) {
      var node = _ref4.node;

      if (node.descr.type === "FuncImportDescr") {
        // $FlowIgnore
        var indexBasedFunctionName = node.descr.id;
        var index = Number(indexBasedFunctionName.replace("func_", ""));
        var functionName = functionNames.find(function (f) {
          return f.index === index;
        ***REMOVED***);

        if (functionName) {
          // $FlowIgnore
          node.descr.id = t.identifier(functionName.name);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***),
    CallInstruction: function (_CallInstruction) {
      function CallInstruction(_x4) {
        return _CallInstruction.apply(this, arguments);
      ***REMOVED***

      CallInstruction.toString = function () {
        return _CallInstruction.toString();
      ***REMOVED***;

      return CallInstruction;
    ***REMOVED***(function (nodePath) {
      var node = nodePath.node;
      var index = node.index.value;
      var functionName = functionNames.find(function (f) {
        return f.index === index;
      ***REMOVED***);

      if (functionName) {
        var oldValue = node.index;
        node.index = t.identifier(functionName.name);
        node.numeric = oldValue; // $FlowIgnore

        delete node.raw;
      ***REMOVED***
    ***REMOVED***)
  ***REMOVED***);
***REMOVED***

function restoreLocalNames(ast) {
  var localNames = [];
  t.traverse(ast, {
    LocalNameMetadata: function LocalNameMetadata(_ref5) {
      var node = _ref5.node;
      localNames.push({
        name: node.value,
        localIndex: node.localIndex,
        functionIndex: node.functionIndex
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***);

  if (localNames.length === 0) {
    return;
  ***REMOVED***

  t.traverse(ast, {
    Func: function (_Func2) {
      function Func(_x5) {
        return _Func2.apply(this, arguments);
      ***REMOVED***

      Func.toString = function () {
        return _Func2.toString();
      ***REMOVED***;

      return Func;
    ***REMOVED***(function (_ref6) {
      var node = _ref6.node;
      var signature = node.signature;

      if (signature.type !== "Signature") {
        return;
      ***REMOVED*** // $FlowIgnore


      var nodeName = node.name;
      var indexBasedFunctionName = nodeName.value;
      var functionIndex = Number(indexBasedFunctionName.replace("func_", ""));
      signature.params.forEach(function (param, paramIndex) {
        var paramName = localNames.find(function (f) {
          return f.localIndex === paramIndex && f.functionIndex === functionIndex;
        ***REMOVED***);

        if (paramName && paramName.name !== "") {
          param.id = paramName.name;
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***)
  ***REMOVED***);
***REMOVED***

function restoreModuleName(ast) {
  t.traverse(ast, {
    ModuleNameMetadata: function (_ModuleNameMetadata) {
      function ModuleNameMetadata(_x6) {
        return _ModuleNameMetadata.apply(this, arguments);
      ***REMOVED***

      ModuleNameMetadata.toString = function () {
        return _ModuleNameMetadata.toString();
      ***REMOVED***;

      return ModuleNameMetadata;
    ***REMOVED***(function (moduleNameMetadataPath) {
      // update module
      t.traverse(ast, {
        Module: function (_Module) {
          function Module(_x7) {
            return _Module.apply(this, arguments);
          ***REMOVED***

          Module.toString = function () {
            return _Module.toString();
          ***REMOVED***;

          return Module;
        ***REMOVED***(function (_ref7) {
          var node = _ref7.node;
          var name = moduleNameMetadataPath.node.value; // compatiblity with wast-parser

          if (name === "") {
            name = null;
          ***REMOVED***

          node.id = name;
        ***REMOVED***)
      ***REMOVED***);
    ***REMOVED***)
  ***REMOVED***);
***REMOVED***

function decode(buf, customOpts) {
  var opts = Object.assign({***REMOVED***, defaultDecoderOpts, customOpts);
  var ast = decoder.decode(buf, opts);

  if (opts.ignoreCustomNameSection === false) {
    restoreFunctionNames(ast);
    restoreLocalNames(ast);
    restoreModuleName(ast);
  ***REMOVED***

  return ast;
***REMOVED***
