// fast-uri/lib/utils.js
'use strict'

const { HEX ***REMOVED*** = require('./scopedChars')

const IPV4_REG = /^(?:(?:25[0-5]|2[0-4]\d|1\d{2***REMOVED***|[1-9]\d|\d)\.){3***REMOVED***(?:25[0-5]|2[0-4]\d|1\d{2***REMOVED***|[1-9]\d|\d)$/u

function normalizeIPv4 (host) {
  if (findToken(host, '.') < 3) { return { host, isIPV4: false ***REMOVED*** ***REMOVED***
  const matches = host.match(IPV4_REG) || []
  const [address] = matches
  if (address) {
    return { host: stripLeadingZeros(address, '.'), isIPV4: true ***REMOVED***
  ***REMOVED*** else {
    return { host, isIPV4: false ***REMOVED***
  ***REMOVED***
***REMOVED***

/**
 * @param {string[]***REMOVED*** input
 * @param {boolean***REMOVED*** [keepZero=false]
 * @returns {string|undefined***REMOVED***
 */
function stringArrayToHexStripped (input, keepZero = false) {
  let acc = ''
  let strip = true
  for (const c of input) {
    if (HEX[c] === undefined) return undefined
    if (c !== '0' && strip === true) strip = false
    if (!strip) acc += c
  ***REMOVED***
  if (keepZero && acc.length === 0) acc = '0'
  return acc
***REMOVED***

function getIPV6 (input) {
  let tokenCount = 0
  const output = { error: false, address: '', zone: '' ***REMOVED***
  const address = []
  const buffer = []
  let isZone = false
  let endipv6Encountered = false
  let endIpv6 = false

  function consume () {
    if (buffer.length) {
      if (isZone === false) {
        const hex = stringArrayToHexStripped(buffer)
        if (hex !== undefined) {
          address.push(hex)
        ***REMOVED*** else {
          output.error = true
          return false
        ***REMOVED***
      ***REMOVED***
      buffer.length = 0
    ***REMOVED***
    return true
  ***REMOVED***

  for (let i = 0; i < input.length; i++) {
    const cursor = input[i]
    if (cursor === '[' || cursor === ']') { continue ***REMOVED***
    if (cursor === ':') {
      if (endipv6Encountered === true) {
        endIpv6 = true
      ***REMOVED***
      if (!consume()) { break ***REMOVED***
      tokenCount++
      address.push(':')
      if (tokenCount > 7) {
        // not valid
        output.error = true
        break
      ***REMOVED***
      if (i - 1 >= 0 && input[i - 1] === ':') {
        endipv6Encountered = true
      ***REMOVED***
      continue
    ***REMOVED*** else if (cursor === '%') {
      if (!consume()) { break ***REMOVED***
      // switch to zone detection
      isZone = true
    ***REMOVED*** else {
      buffer.push(cursor)
      continue
    ***REMOVED***
  ***REMOVED***
  if (buffer.length) {
    if (isZone) {
      output.zone = buffer.join('')
    ***REMOVED*** else if (endIpv6) {
      address.push(buffer.join(''))
    ***REMOVED*** else {
      address.push(stringArrayToHexStripped(buffer))
    ***REMOVED***
  ***REMOVED***
  output.address = address.join('')
  return output
***REMOVED***

function normalizeIPv6 (host) {
  if (findToken(host, ':') < 2) { return { host, isIPV6: false ***REMOVED*** ***REMOVED***
  const ipv6 = getIPV6(host)

  if (!ipv6.error) {
    let newHost = ipv6.address
    let escapedHost = ipv6.address
    if (ipv6.zone) {
      newHost += '%' + ipv6.zone
      escapedHost += '%25' + ipv6.zone
    ***REMOVED***
    return { host: newHost, escapedHost, isIPV6: true ***REMOVED***
  ***REMOVED*** else {
    return { host, isIPV6: false ***REMOVED***
  ***REMOVED***
***REMOVED***

function stripLeadingZeros (str, token) {
  let out = ''
  let skip = true
  const l = str.length
  for (let i = 0; i < l; i++) {
    const c = str[i]
    if (c === '0' && skip) {
      if ((i + 1 <= l && str[i + 1] === token) || i + 1 === l) {
        out += c
        skip = false
      ***REMOVED***
    ***REMOVED*** else {
      if (c === token) {
        skip = true
      ***REMOVED*** else {
        skip = false
      ***REMOVED***
      out += c
    ***REMOVED***
  ***REMOVED***
  return out
***REMOVED***

function findToken (str, token) {
  let ind = 0
  for (let i = 0; i < str.length; i++) {
    if (str[i] === token) ind++
  ***REMOVED***
  return ind
***REMOVED***

const RDS1 = /^\.\.?\//u
const RDS2 = /^\/\.(?:\/|$)/u
const RDS3 = /^\/\.\.(?:\/|$)/u
const RDS5 = /^\/?(?:.|\n)*?(?=\/|$)/u

function removeDotSegments (input) {
  const output = []

  while (input.length) {
    if (input.match(RDS1)) {
      input = input.replace(RDS1, '')
    ***REMOVED*** else if (input.match(RDS2)) {
      input = input.replace(RDS2, '/')
    ***REMOVED*** else if (input.match(RDS3)) {
      input = input.replace(RDS3, '/')
      output.pop()
    ***REMOVED*** else if (input === '.' || input === '..') {
      input = ''
    ***REMOVED*** else {
      const im = input.match(RDS5)
      if (im) {
        const s = im[0]
        input = input.slice(s.length)
        output.push(s)
      ***REMOVED*** else {
        throw new Error('Unexpected dot segment condition')
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  return output.join('')
***REMOVED***

function normalizeComponentEncoding (components, esc) {
  const func = esc !== true ? escape : unescape
  if (components.scheme !== undefined) {
    components.scheme = func(components.scheme)
  ***REMOVED***
  if (components.userinfo !== undefined) {
    components.userinfo = func(components.userinfo)
  ***REMOVED***
  if (components.host !== undefined) {
    components.host = func(components.host)
  ***REMOVED***
  if (components.path !== undefined) {
    components.path = func(components.path)
  ***REMOVED***
  if (components.query !== undefined) {
    components.query = func(components.query)
  ***REMOVED***
  if (components.fragment !== undefined) {
    components.fragment = func(components.fragment)
  ***REMOVED***
  return components
***REMOVED***

function recomposeAuthority (components) {
  const uriTokens = []

  if (components.userinfo !== undefined) {
    uriTokens.push(components.userinfo)
    uriTokens.push('@')
  ***REMOVED***

  if (components.host !== undefined) {
    let host = unescape(components.host)
    const ipV4res = normalizeIPv4(host)

    if (ipV4res.isIPV4) {
      host = ipV4res.host
    ***REMOVED*** else {
      const ipV6res = normalizeIPv6(ipV4res.host)
      if (ipV6res.isIPV6 === true) {
        host = `[${ipV6res.escapedHost***REMOVED***]`
      ***REMOVED*** else {
        host = components.host
      ***REMOVED***
    ***REMOVED***
    uriTokens.push(host)
  ***REMOVED***

  if (typeof components.port === 'number' || typeof components.port === 'string') {
    uriTokens.push(':')
    uriTokens.push(String(components.port))
  ***REMOVED***

  return uriTokens.length ? uriTokens.join('') : undefined
***REMOVED***;

module.exports = {
  recomposeAuthority,
  normalizeComponentEncoding,
  removeDotSegments,
  normalizeIPv4,
  normalizeIPv6,
  stringArrayToHexStripped
***REMOVED***
