// eslint/lib/init/npm-utils.js
/**
 * @fileoverview Utility for executing npm commands.
 * @author Ian VanSchooten
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const fs = require("fs"),
    spawn = require("cross-spawn"),
    path = require("path"),
    log = require("../shared/logging");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Find the closest package.json file, starting at process.cwd (by default),
 * and working up to root.
 * @param   {string***REMOVED*** [startDir=process.cwd()] Starting directory
 * @returns {string***REMOVED***                          Absolute path to closest package.json file
 */
function findPackageJson(startDir) {
    let dir = path.resolve(startDir || process.cwd());

    do {
        const pkgFile = path.join(dir, "package.json");

        if (!fs.existsSync(pkgFile) || !fs.statSync(pkgFile).isFile()) {
            dir = path.join(dir, "..");
            continue;
        ***REMOVED***
        return pkgFile;
    ***REMOVED*** while (dir !== path.resolve(dir, ".."));
    return null;
***REMOVED***

//------------------------------------------------------------------------------
// Private
//------------------------------------------------------------------------------

/**
 * Install node modules synchronously and save to devDependencies in package.json
 * @param   {string|string[]***REMOVED*** packages Node module or modules to install
 * @returns {void***REMOVED***
 */
function installSyncSaveDev(packages) {
    const packageList = Array.isArray(packages) ? packages : [packages];
    const npmProcess = spawn.sync("npm", ["i", "--save-dev"].concat(packageList), { stdio: "inherit" ***REMOVED***);
    const error = npmProcess.error;

    if (error && error.code === "ENOENT") {
        const pluralS = packageList.length > 1 ? "s" : "";

        log.error(`Could not execute npm. Please install the following package${pluralS***REMOVED*** with a package manager of your choice: ${packageList.join(", ")***REMOVED***`);
    ***REMOVED***
***REMOVED***

/**
 * Fetch `peerDependencies` of the given package by `npm show` command.
 * @param {string***REMOVED*** packageName The package name to fetch peerDependencies.
 * @returns {Object***REMOVED*** Gotten peerDependencies. Returns null if npm was not found.
 */
function fetchPeerDependencies(packageName) {
    const npmProcess = spawn.sync(
        "npm",
        ["show", "--json", packageName, "peerDependencies"],
        { encoding: "utf8" ***REMOVED***
    );

    const error = npmProcess.error;

    if (error && error.code === "ENOENT") {
        return null;
    ***REMOVED***
    const fetchedText = npmProcess.stdout.trim();

    return JSON.parse(fetchedText || "{***REMOVED***");


***REMOVED***

/**
 * Check whether node modules are include in a project's package.json.
 * @param   {string[]***REMOVED*** packages           Array of node module names
 * @param   {Object***REMOVED***  opt                 Options Object
 * @param   {boolean***REMOVED*** opt.dependencies    Set to true to check for direct dependencies
 * @param   {boolean***REMOVED*** opt.devDependencies Set to true to check for development dependencies
 * @param   {boolean***REMOVED*** opt.startdir        Directory to begin searching from
 * @returns {Object***REMOVED***                      An object whose keys are the module names
 *                                        and values are booleans indicating installation.
 */
function check(packages, opt) {
    const deps = new Set();
    const pkgJson = (opt) ? findPackageJson(opt.startDir) : findPackageJson();
    let fileJson;

    if (!pkgJson) {
        throw new Error("Could not find a package.json file. Run 'npm init' to create one.");
    ***REMOVED***

    try {
        fileJson = JSON.parse(fs.readFileSync(pkgJson, "utf8"));
    ***REMOVED*** catch (e) {
        const error = new Error(e);

        error.messageTemplate = "failed-to-read-json";
        error.messageData = {
            path: pkgJson,
            message: e.message
        ***REMOVED***;
        throw error;
    ***REMOVED***

    ["dependencies", "devDependencies"].forEach(key => {
        if (opt[key] && typeof fileJson[key] === "object") {
            Object.keys(fileJson[key]).forEach(dep => deps.add(dep));
        ***REMOVED***
    ***REMOVED***);

    return packages.reduce((status, pkg) => {
        status[pkg] = deps.has(pkg);
        return status;
    ***REMOVED***, {***REMOVED***);
***REMOVED***

/**
 * Check whether node modules are included in the dependencies of a project's
 * package.json.
 *
 * Convenience wrapper around check().
 * @param   {string[]***REMOVED*** packages  Array of node modules to check.
 * @param   {string***REMOVED***   rootDir   The directory containing a package.json
 * @returns {Object***REMOVED***             An object whose keys are the module names
 *                               and values are booleans indicating installation.
 */
function checkDeps(packages, rootDir) {
    return check(packages, { dependencies: true, startDir: rootDir ***REMOVED***);
***REMOVED***

/**
 * Check whether node modules are included in the devDependencies of a project's
 * package.json.
 *
 * Convenience wrapper around check().
 * @param   {string[]***REMOVED*** packages  Array of node modules to check.
 * @returns {Object***REMOVED***             An object whose keys are the module names
 *                               and values are booleans indicating installation.
 */
function checkDevDeps(packages) {
    return check(packages, { devDependencies: true ***REMOVED***);
***REMOVED***

/**
 * Check whether package.json is found in current path.
 * @param   {string***REMOVED*** [startDir] Starting directory
 * @returns {boolean***REMOVED*** Whether a package.json is found in current path.
 */
function checkPackageJson(startDir) {
    return !!findPackageJson(startDir);
***REMOVED***

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

module.exports = {
    installSyncSaveDev,
    fetchPeerDependencies,
    findPackageJson,
    checkDeps,
    checkDevDeps,
    checkPackageJson
***REMOVED***;
