// jsx-ast-utils/__tests__/helper.js
/* eslint-env jest */
import getProp from '../src/getProp';

const nodeVersion = parseInt(process.version.match(/^v(\d+)\./)[1], 10);

export const fallbackToBabylon = nodeVersion < 6;

let parserName;
const babelParser = fallbackToBabylon ? require('babylon') : require('@babel/parser');
const flowParser = require('flow-parser');

const defaultPlugins = [
  'jsx',
  'functionBind',
  'estree',
  'objectRestSpread',
  'optionalChaining',
  // 'nullishCoalescing', // TODO: update to babel 7
];
let plugins = [...defaultPlugins];
let isESM = false;

export function setParserName(name) {
  parserName = name;
***REMOVED***

export function changePlugins(pluginOrFn) {
  if (Array.isArray(pluginOrFn)) {
    plugins = pluginOrFn;
  ***REMOVED*** else if (typeof pluginOrFn === 'function') {
    plugins = pluginOrFn(plugins);
  ***REMOVED*** else {
    throw new Error('changePlugins argument should be either an array or a function');
  ***REMOVED***
***REMOVED***

export function setIsESM() {
  isESM = true;
***REMOVED***

beforeEach(() => {
  plugins = [...defaultPlugins];
  isESM = false;
***REMOVED***);

function parse(code) {
  if (parserName === undefined) {
    throw new Error('No parser specified');
  ***REMOVED***
  if (parserName === 'babel') {
    try {
      return babelParser.parse(code, { plugins, sourceFilename: 'test.js', ...(isESM && { sourceType: 'module' ***REMOVED***) ***REMOVED***);
    ***REMOVED*** catch (_) {
      // eslint-disable-next-line no-console
      console.warn(`Failed to parse with ${fallbackToBabylon ? 'babylon' : 'Babel'***REMOVED*** parser.`);
    ***REMOVED***
  ***REMOVED***
  if (parserName === 'flow') {
    try {
      return flowParser.parse(code, { plugins ***REMOVED***);
    ***REMOVED*** catch (_) {
      // eslint-disable-next-line no-console
      console.warn('Failed to parse with the Flow parser');
    ***REMOVED***
  ***REMOVED***
  throw new Error(`The parser ${parserName***REMOVED*** is not yet supported for testing.`);
***REMOVED***

export function getOpeningElement(code) {
  const parsedCode = parse(code);
  let body;
  if (parsedCode.program) {
    // eslint-disable-next-line prefer-destructuring
    body = parsedCode.program.body;
  ***REMOVED*** else {
    // eslint-disable-next-line prefer-destructuring
    body = parsedCode.body;
  ***REMOVED***
  if (Array.isArray(body) && body[0] != null) {
    const [{ expression ***REMOVED***] = body;
    return expression.type === 'JSXFragment' ? expression.openingFragment : expression.openingElement;
  ***REMOVED***

  return null;
***REMOVED***

export function extractProp(code, prop = 'foo') {
  const node = getOpeningElement(code);
  const { attributes: props ***REMOVED*** = node;
  return getProp(props, prop);
***REMOVED***

export const describeIfNotBabylon = fallbackToBabylon ? describe.skip : describe;
