// lodash.sortby/index.js
/**
 * lodash (Custom Build) <https://lodash.com/>
 * Build: `lodash modularize exports="npm" -o ./`
 * Copyright jQuery Foundation and other contributors <https://jquery.org/>
 * Released under MIT license <https://lodash.com/license>
 * Based on Underscore.js 1.8.3 <http://underscorejs.org/LICENSE>
 * Copyright Jeremy Ashkenas, DocumentCloud and Investigative Reporters & Editors
 */

/** Used as the size to enable large array optimizations. */
var LARGE_ARRAY_SIZE = 200;

/** Used as the `TypeError` message for "Functions" methods. */
var FUNC_ERROR_TEXT = 'Expected a function';

/** Used to stand-in for `undefined` hash values. */
var HASH_UNDEFINED = '__lodash_hash_undefined__';

/** Used to compose bitmasks for comparison styles. */
var UNORDERED_COMPARE_FLAG = 1,
    PARTIAL_COMPARE_FLAG = 2;

/** Used as references for various `Number` constants. */
var INFINITY = 1 / 0,
    MAX_SAFE_INTEGER = 9007199254740991;

/** `Object#toString` result references. */
var argsTag = '[object Arguments]',
    arrayTag = '[object Array]',
    boolTag = '[object Boolean]',
    dateTag = '[object Date]',
    errorTag = '[object Error]',
    funcTag = '[object Function]',
    genTag = '[object GeneratorFunction]',
    mapTag = '[object Map]',
    numberTag = '[object Number]',
    objectTag = '[object Object]',
    promiseTag = '[object Promise]',
    regexpTag = '[object RegExp]',
    setTag = '[object Set]',
    stringTag = '[object String]',
    symbolTag = '[object Symbol]',
    weakMapTag = '[object WeakMap]';

var arrayBufferTag = '[object ArrayBuffer]',
    dataViewTag = '[object DataView]',
    float32Tag = '[object Float32Array]',
    float64Tag = '[object Float64Array]',
    int8Tag = '[object Int8Array]',
    int16Tag = '[object Int16Array]',
    int32Tag = '[object Int32Array]',
    uint8Tag = '[object Uint8Array]',
    uint8ClampedTag = '[object Uint8ClampedArray]',
    uint16Tag = '[object Uint16Array]',
    uint32Tag = '[object Uint32Array]';

/** Used to match property names within property paths. */
var reIsDeepProp = /\.|\[(?:[^[\]]*|(["'])(?:(?!\1)[^\\]|\\.)*?\1)\]/,
    reIsPlainProp = /^\w*$/,
    reLeadingDot = /^\./,
    rePropName = /[^.[\]]+|\[(?:(-?\d+(?:\.\d+)?)|(["'])((?:(?!\2)[^\\]|\\.)*?)\2)\]|(?=(?:\.|\[\])(?:\.|\[\]|$))/g;

/**
 * Used to match `RegExp`
 * [syntax characters](http://ecma-international.org/ecma-262/7.0/#sec-patterns).
 */
var reRegExpChar = /[\\^$.*+?()[\]{***REMOVED***|]/g;

/** Used to match backslashes in property paths. */
var reEscapeChar = /\\(\\)?/g;

/** Used to detect host constructors (Safari). */
var reIsHostCtor = /^\[object .+?Constructor\]$/;

/** Used to detect unsigned integer values. */
var reIsUint = /^(?:0|[1-9]\d*)$/;

/** Used to identify `toStringTag` values of typed arrays. */
var typedArrayTags = {***REMOVED***;
typedArrayTags[float32Tag] = typedArrayTags[float64Tag] =
typedArrayTags[int8Tag] = typedArrayTags[int16Tag] =
typedArrayTags[int32Tag] = typedArrayTags[uint8Tag] =
typedArrayTags[uint8ClampedTag] = typedArrayTags[uint16Tag] =
typedArrayTags[uint32Tag] = true;
typedArrayTags[argsTag] = typedArrayTags[arrayTag] =
typedArrayTags[arrayBufferTag] = typedArrayTags[boolTag] =
typedArrayTags[dataViewTag] = typedArrayTags[dateTag] =
typedArrayTags[errorTag] = typedArrayTags[funcTag] =
typedArrayTags[mapTag] = typedArrayTags[numberTag] =
typedArrayTags[objectTag] = typedArrayTags[regexpTag] =
typedArrayTags[setTag] = typedArrayTags[stringTag] =
typedArrayTags[weakMapTag] = false;

/** Detect free variable `global` from Node.js. */
var freeGlobal = typeof global == 'object' && global && global.Object === Object && global;

/** Detect free variable `self`. */
var freeSelf = typeof self == 'object' && self && self.Object === Object && self;

/** Used as a reference to the global object. */
var root = freeGlobal || freeSelf || Function('return this')();

/** Detect free variable `exports`. */
var freeExports = typeof exports == 'object' && exports && !exports.nodeType && exports;

/** Detect free variable `module`. */
var freeModule = freeExports && typeof module == 'object' && module && !module.nodeType && module;

/** Detect the popular CommonJS extension `module.exports`. */
var moduleExports = freeModule && freeModule.exports === freeExports;

/** Detect free variable `process` from Node.js. */
var freeProcess = moduleExports && freeGlobal.process;

/** Used to access faster Node.js helpers. */
var nodeUtil = (function() {
  try {
    return freeProcess && freeProcess.binding('util');
  ***REMOVED*** catch (e) {***REMOVED***
***REMOVED***());

/* Node.js helper references. */
var nodeIsTypedArray = nodeUtil && nodeUtil.isTypedArray;

/**
 * A faster alternative to `Function#apply`, this function invokes `func`
 * with the `this` binding of `thisArg` and the arguments of `args`.
 *
 * @private
 * @param {Function***REMOVED*** func The function to invoke.
 * @param {****REMOVED*** thisArg The `this` binding of `func`.
 * @param {Array***REMOVED*** args The arguments to invoke `func` with.
 * @returns {****REMOVED*** Returns the result of `func`.
 */
function apply(func, thisArg, args) {
  switch (args.length) {
    case 0: return func.call(thisArg);
    case 1: return func.call(thisArg, args[0]);
    case 2: return func.call(thisArg, args[0], args[1]);
    case 3: return func.call(thisArg, args[0], args[1], args[2]);
  ***REMOVED***
  return func.apply(thisArg, args);
***REMOVED***

/**
 * A specialized version of `_.map` for arrays without support for iteratee
 * shorthands.
 *
 * @private
 * @param {Array***REMOVED*** [array] The array to iterate over.
 * @param {Function***REMOVED*** iteratee The function invoked per iteration.
 * @returns {Array***REMOVED*** Returns the new mapped array.
 */
function arrayMap(array, iteratee) {
  var index = -1,
      length = array ? array.length : 0,
      result = Array(length);

  while (++index < length) {
    result[index] = iteratee(array[index], index, array);
  ***REMOVED***
  return result;
***REMOVED***

/**
 * Appends the elements of `values` to `array`.
 *
 * @private
 * @param {Array***REMOVED*** array The array to modify.
 * @param {Array***REMOVED*** values The values to append.
 * @returns {Array***REMOVED*** Returns `array`.
 */
function arrayPush(array, values) {
  var index = -1,
      length = values.length,
      offset = array.length;

  while (++index < length) {
    array[offset + index] = values[index];
  ***REMOVED***
  return array;
***REMOVED***

/**
 * A specialized version of `_.some` for arrays without support for iteratee
 * shorthands.
 *
 * @private
 * @param {Array***REMOVED*** [array] The array to iterate over.
 * @param {Function***REMOVED*** predicate The function invoked per iteration.
 * @returns {boolean***REMOVED*** Returns `true` if any element passes the predicate check,
 *  else `false`.
 */
function arraySome(array, predicate) {
  var index = -1,
      length = array ? array.length : 0;

  while (++index < length) {
    if (predicate(array[index], index, array)) {
      return true;
    ***REMOVED***
  ***REMOVED***
  return false;
***REMOVED***

/**
 * The base implementation of `_.property` without support for deep paths.
 *
 * @private
 * @param {string***REMOVED*** key The key of the property to get.
 * @returns {Function***REMOVED*** Returns the new accessor function.
 */
function baseProperty(key) {
  return function(object) {
    return object == null ? undefined : object[key];
  ***REMOVED***;
***REMOVED***

/**
 * The base implementation of `_.sortBy` which uses `comparer` to define the
 * sort order of `array` and replaces criteria objects with their corresponding
 * values.
 *
 * @private
 * @param {Array***REMOVED*** array The array to sort.
 * @param {Function***REMOVED*** comparer The function to define sort order.
 * @returns {Array***REMOVED*** Returns `array`.
 */
function baseSortBy(array, comparer) {
  var length = array.length;

  array.sort(comparer);
  while (length--) {
    array[length] = array[length].value;
  ***REMOVED***
  return array;
***REMOVED***

/**
 * The base implementation of `_.times` without support for iteratee shorthands
 * or max array length checks.
 *
 * @private
 * @param {number***REMOVED*** n The number of times to invoke `iteratee`.
 * @param {Function***REMOVED*** iteratee The function invoked per iteration.
 * @returns {Array***REMOVED*** Returns the array of results.
 */
function baseTimes(n, iteratee) {
  var index = -1,
      result = Array(n);

  while (++index < n) {
    result[index] = iteratee(index);
  ***REMOVED***
  return result;
***REMOVED***

/**
 * The base implementation of `_.unary` without support for storing metadata.
 *
 * @private
 * @param {Function***REMOVED*** func The function to cap arguments for.
 * @returns {Function***REMOVED*** Returns the new capped function.
 */
function baseUnary(func) {
  return function(value) {
    return func(value);
  ***REMOVED***;
***REMOVED***

/**
 * Gets the value at `key` of `object`.
 *
 * @private
 * @param {Object***REMOVED*** [object] The object to query.
 * @param {string***REMOVED*** key The key of the property to get.
 * @returns {****REMOVED*** Returns the property value.
 */
function getValue(object, key) {
  return object == null ? undefined : object[key];
***REMOVED***

/**
 * Checks if `value` is a host object in IE < 9.
 *
 * @private
 * @param {****REMOVED*** value The value to check.
 * @returns {boolean***REMOVED*** Returns `true` if `value` is a host object, else `false`.
 */
function isHostObject(value) {
  // Many host objects are `Object` objects that can coerce to strings
  // despite having improperly defined `toString` methods.
  var result = false;
  if (value != null && typeof value.toString != 'function') {
    try {
      result = !!(value + '');
    ***REMOVED*** catch (e) {***REMOVED***
  ***REMOVED***
  return result;
***REMOVED***

/**
 * Converts `map` to its key-value pairs.
 *
 * @private
 * @param {Object***REMOVED*** map The map to convert.
 * @returns {Array***REMOVED*** Returns the key-value pairs.
 */
function mapToArray(map) {
  var index = -1,
      result = Array(map.size);

  map.forEach(function(value, key) {
    result[++index] = [key, value];
  ***REMOVED***);
  return result;
***REMOVED***

/**
 * Creates a unary function that invokes `func` with its argument transformed.
 *
 * @private
 * @param {Function***REMOVED*** func The function to wrap.
 * @param {Function***REMOVED*** transform The argument transform.
 * @returns {Function***REMOVED*** Returns the new function.
 */
function overArg(func, transform) {
  return function(arg) {
    return func(transform(arg));
  ***REMOVED***;
***REMOVED***

/**
 * Converts `set` to an array of its values.
 *
 * @private
 * @param {Object***REMOVED*** set The set to convert.
 * @returns {Array***REMOVED*** Returns the values.
 */
function setToArray(set) {
  var index = -1,
      result = Array(set.size);

  set.forEach(function(value) {
    result[++index] = value;
  ***REMOVED***);
  return result;
***REMOVED***

/** Used for built-in method references. */
var arrayProto = Array.prototype,
    funcProto = Function.prototype,
    objectProto = Object.prototype;

/** Used to detect overreaching core-js shims. */
var coreJsData = root['__core-js_shared__'];

/** Used to detect methods masquerading as native. */
var maskSrcKey = (function() {
  var uid = /[^.]+$/.exec(coreJsData && coreJsData.keys && coreJsData.keys.IE_PROTO || '');
  return uid ? ('Symbol(src)_1.' + uid) : '';
***REMOVED***());

/** Used to resolve the decompiled source of functions. */
var funcToString = funcProto.toString;

/** Used to check objects for own properties. */
var hasOwnProperty = objectProto.hasOwnProperty;

/**
 * Used to resolve the
 * [`toStringTag`](http://ecma-international.org/ecma-262/7.0/#sec-object.prototype.tostring)
 * of values.
 */
var objectToString = objectProto.toString;

/** Used to detect if a method is native. */
var reIsNative = RegExp('^' +
  funcToString.call(hasOwnProperty).replace(reRegExpChar, '\\$&')
  .replace(/hasOwnProperty|(function).*?(?=\\\()| for .+?(?=\\\])/g, '$1.*?') + '$'
);

/** Built-in value references. */
var Symbol = root.Symbol,
    Uint8Array = root.Uint8Array,
    propertyIsEnumerable = objectProto.propertyIsEnumerable,
    splice = arrayProto.splice,
    spreadableSymbol = Symbol ? Symbol.isConcatSpreadable : undefined;

/* Built-in method references for those with the same name as other `lodash` methods. */
var nativeKeys = overArg(Object.keys, Object),
    nativeMax = Math.max;

/* Built-in method references that are verified to be native. */
var DataView = getNative(root, 'DataView'),
    Map = getNative(root, 'Map'),
    Promise = getNative(root, 'Promise'),
    Set = getNative(root, 'Set'),
    WeakMap = getNative(root, 'WeakMap'),
    nativeCreate = getNative(Object, 'create');

/** Used to detect maps, sets, and weakmaps. */
var dataViewCtorString = toSource(DataView),
    mapCtorString = toSource(Map),
    promiseCtorString = toSource(Promise),
    setCtorString = toSource(Set),
    weakMapCtorString = toSource(WeakMap);

/** Used to convert symbols to primitives and strings. */
var symbolProto = Symbol ? Symbol.prototype : undefined,
    symbolValueOf = symbolProto ? symbolProto.valueOf : undefined,
    symbolToString = symbolProto ? symbolProto.toString : undefined;

/**
 * Creates a hash object.
 *
 * @private
 * @constructor
 * @param {Array***REMOVED*** [entries] The key-value pairs to cache.
 */
function Hash(entries) {
  var index = -1,
      length = entries ? entries.length : 0;

  this.clear();
  while (++index < length) {
    var entry = entries[index];
    this.set(entry[0], entry[1]);
  ***REMOVED***
***REMOVED***

/**
 * Removes all key-value entries from the hash.
 *
 * @private
 * @name clear
 * @memberOf Hash
 */
function hashClear() {
  this.__data__ = nativeCreate ? nativeCreate(null) : {***REMOVED***;
***REMOVED***

/**
 * Removes `key` and its value from the hash.
 *
 * @private
 * @name delete
 * @memberOf Hash
 * @param {Object***REMOVED*** hash The hash to modify.
 * @param {string***REMOVED*** key The key of the value to remove.
 * @returns {boolean***REMOVED*** Returns `true` if the entry was removed, else `false`.
 */
function hashDelete(key) {
  return this.has(key) && delete this.__data__[key];
***REMOVED***

/**
 * Gets the hash value for `key`.
 *
 * @private
 * @name get
 * @memberOf Hash
 * @param {string***REMOVED*** key The key of the value to get.
 * @returns {****REMOVED*** Returns the entry value.
 */
function hashGet(key) {
  var data = this.__data__;
  if (nativeCreate) {
    var result = data[key];
    return result === HASH_UNDEFINED ? undefined : result;
  ***REMOVED***
  return hasOwnProperty.call(data, key) ? data[key] : undefined;
***REMOVED***

/**
 * Checks if a hash value for `key` exists.
 *
 * @private
 * @name has
 * @memberOf Hash
 * @param {string***REMOVED*** key The key of the entry to check.
 * @returns {boolean***REMOVED*** Returns `true` if an entry for `key` exists, else `false`.
 */
function hashHas(key) {
  var data = this.__data__;
  return nativeCreate ? data[key] !== undefined : hasOwnProperty.call(data, key);
***REMOVED***

/**
 * Sets the hash `key` to `value`.
 *
 * @private
 * @name set
 * @memberOf Hash
 * @param {string***REMOVED*** key The key of the value to set.
 * @param {****REMOVED*** value The value to set.
 * @returns {Object***REMOVED*** Returns the hash instance.
 */
function hashSet(key, value) {
  var data = this.__data__;
  data[key] = (nativeCreate && value === undefined) ? HASH_UNDEFINED : value;
  return this;
***REMOVED***

// Add methods to `Hash`.
Hash.prototype.clear = hashClear;
Hash.prototype['delete'] = hashDelete;
Hash.prototype.get = hashGet;
Hash.prototype.has = hashHas;
Hash.prototype.set = hashSet;

/**
 * Creates an list cache object.
 *
 * @private
 * @constructor
 * @param {Array***REMOVED*** [entries] The key-value pairs to cache.
 */
function ListCache(entries) {
  var index = -1,
      length = entries ? entries.length : 0;

  this.clear();
  while (++index < length) {
    var entry = entries[index];
    this.set(entry[0], entry[1]);
  ***REMOVED***
***REMOVED***

/**
 * Removes all key-value entries from the list cache.
 *
 * @private
 * @name clear
 * @memberOf ListCache
 */
function listCacheClear() {
  this.__data__ = [];
***REMOVED***

/**
 * Removes `key` and its value from the list cache.
 *
 * @private
 * @name delete
 * @memberOf ListCache
 * @param {string***REMOVED*** key The key of the value to remove.
 * @returns {boolean***REMOVED*** Returns `true` if the entry was removed, else `false`.
 */
function listCacheDelete(key) {
  var data = this.__data__,
      index = assocIndexOf(data, key);

  if (index < 0) {
    return false;
  ***REMOVED***
  var lastIndex = data.length - 1;
  if (index == lastIndex) {
    data.pop();
  ***REMOVED*** else {
    splice.call(data, index, 1);
  ***REMOVED***
  return true;
***REMOVED***

/**
 * Gets the list cache value for `key`.
 *
 * @private
 * @name get
 * @memberOf ListCache
 * @param {string***REMOVED*** key The key of the value to get.
 * @returns {****REMOVED*** Returns the entry value.
 */
function listCacheGet(key) {
  var data = this.__data__,
      index = assocIndexOf(data, key);

  return index < 0 ? undefined : data[index][1];
***REMOVED***

/**
 * Checks if a list cache value for `key` exists.
 *
 * @private
 * @name has
 * @memberOf ListCache
 * @param {string***REMOVED*** key The key of the entry to check.
 * @returns {boolean***REMOVED*** Returns `true` if an entry for `key` exists, else `false`.
 */
function listCacheHas(key) {
  return assocIndexOf(this.__data__, key) > -1;
***REMOVED***

/**
 * Sets the list cache `key` to `value`.
 *
 * @private
 * @name set
 * @memberOf ListCache
 * @param {string***REMOVED*** key The key of the value to set.
 * @param {****REMOVED*** value The value to set.
 * @returns {Object***REMOVED*** Returns the list cache instance.
 */
function listCacheSet(key, value) {
  var data = this.__data__,
      index = assocIndexOf(data, key);

  if (index < 0) {
    data.push([key, value]);
  ***REMOVED*** else {
    data[index][1] = value;
  ***REMOVED***
  return this;
***REMOVED***

// Add methods to `ListCache`.
ListCache.prototype.clear = listCacheClear;
ListCache.prototype['delete'] = listCacheDelete;
ListCache.prototype.get = listCacheGet;
ListCache.prototype.has = listCacheHas;
ListCache.prototype.set = listCacheSet;

/**
 * Creates a map cache object to store key-value pairs.
 *
 * @private
 * @constructor
 * @param {Array***REMOVED*** [entries] The key-value pairs to cache.
 */
function MapCache(entries) {
  var index = -1,
      length = entries ? entries.length : 0;

  this.clear();
  while (++index < length) {
    var entry = entries[index];
    this.set(entry[0], entry[1]);
  ***REMOVED***
***REMOVED***

/**
 * Removes all key-value entries from the map.
 *
 * @private
 * @name clear
 * @memberOf MapCache
 */
function mapCacheClear() {
  this.__data__ = {
    'hash': new Hash,
    'map': new (Map || ListCache),
    'string': new Hash
  ***REMOVED***;
***REMOVED***

/**
 * Removes `key` and its value from the map.
 *
 * @private
 * @name delete
 * @memberOf MapCache
 * @param {string***REMOVED*** key The key of the value to remove.
 * @returns {boolean***REMOVED*** Returns `true` if the entry was removed, else `false`.
 */
function mapCacheDelete(key) {
  return getMapData(this, key)['delete'](key);
***REMOVED***

/**
 * Gets the map value for `key`.
 *
 * @private
 * @name get
 * @memberOf MapCache
 * @param {string***REMOVED*** key The key of the value to get.
 * @returns {****REMOVED*** Returns the entry value.
 */
function mapCacheGet(key) {
  return getMapData(this, key).get(key);
***REMOVED***

/**
 * Checks if a map value for `key` exists.
 *
 * @private
 * @name has
 * @memberOf MapCache
 * @param {string***REMOVED*** key The key of the entry to check.
 * @returns {boolean***REMOVED*** Returns `true` if an entry for `key` exists, else `false`.
 */
function mapCacheHas(key) {
  return getMapData(this, key).has(key);
***REMOVED***

/**
 * Sets the map `key` to `value`.
 *
 * @private
 * @name set
 * @memberOf MapCache
 * @param {string***REMOVED*** key The key of the value to set.
 * @param {****REMOVED*** value The value to set.
 * @returns {Object***REMOVED*** Returns the map cache instance.
 */
function mapCacheSet(key, value) {
  getMapData(this, key).set(key, value);
  return this;
***REMOVED***

// Add methods to `MapCache`.
MapCache.prototype.clear = mapCacheClear;
MapCache.prototype['delete'] = mapCacheDelete;
MapCache.prototype.get = mapCacheGet;
MapCache.prototype.has = mapCacheHas;
MapCache.prototype.set = mapCacheSet;

/**
 *
 * Creates an array cache object to store unique values.
 *
 * @private
 * @constructor
 * @param {Array***REMOVED*** [values] The values to cache.
 */
function SetCache(values) {
  var index = -1,
      length = values ? values.length : 0;

  this.__data__ = new MapCache;
  while (++index < length) {
    this.add(values[index]);
  ***REMOVED***
***REMOVED***

/**
 * Adds `value` to the array cache.
 *
 * @private
 * @name add
 * @memberOf SetCache
 * @alias push
 * @param {****REMOVED*** value The value to cache.
 * @returns {Object***REMOVED*** Returns the cache instance.
 */
function setCacheAdd(value) {
  this.__data__.set(value, HASH_UNDEFINED);
  return this;
***REMOVED***

/**
 * Checks if `value` is in the array cache.
 *
 * @private
 * @name has
 * @memberOf SetCache
 * @param {****REMOVED*** value The value to search for.
 * @returns {number***REMOVED*** Returns `true` if `value` is found, else `false`.
 */
function setCacheHas(value) {
  return this.__data__.has(value);
***REMOVED***

// Add methods to `SetCache`.
SetCache.prototype.add = SetCache.prototype.push = setCacheAdd;
SetCache.prototype.has = setCacheHas;

/**
 * Creates a stack cache object to store key-value pairs.
 *
 * @private
 * @constructor
 * @param {Array***REMOVED*** [entries] The key-value pairs to cache.
 */
function Stack(entries) {
  this.__data__ = new ListCache(entries);
***REMOVED***

/**
 * Removes all key-value entries from the stack.
 *
 * @private
 * @name clear
 * @memberOf Stack
 */
function stackClear() {
  this.__data__ = new ListCache;
***REMOVED***

/**
 * Removes `key` and its value from the stack.
 *
 * @private
 * @name delete
 * @memberOf Stack
 * @param {string***REMOVED*** key The key of the value to remove.
 * @returns {boolean***REMOVED*** Returns `true` if the entry was removed, else `false`.
 */
function stackDelete(key) {
  return this.__data__['delete'](key);
***REMOVED***

/**
 * Gets the stack value for `key`.
 *
 * @private
 * @name get
 * @memberOf Stack
 * @param {string***REMOVED*** key The key of the value to get.
 * @returns {****REMOVED*** Returns the entry value.
 */
function stackGet(key) {
  return this.__data__.get(key);
***REMOVED***

/**
 * Checks if a stack value for `key` exists.
 *
 * @private
 * @name has
 * @memberOf Stack
 * @param {string***REMOVED*** key The key of the entry to check.
 * @returns {boolean***REMOVED*** Returns `true` if an entry for `key` exists, else `false`.
 */
function stackHas(key) {
  return this.__data__.has(key);
***REMOVED***

/**
 * Sets the stack `key` to `value`.
 *
 * @private
 * @name set
 * @memberOf Stack
 * @param {string***REMOVED*** key The key of the value to set.
 * @param {****REMOVED*** value The value to set.
 * @returns {Object***REMOVED*** Returns the stack cache instance.
 */
function stackSet(key, value) {
  var cache = this.__data__;
  if (cache instanceof ListCache) {
    var pairs = cache.__data__;
    if (!Map || (pairs.length < LARGE_ARRAY_SIZE - 1)) {
      pairs.push([key, value]);
      return this;
    ***REMOVED***
    cache = this.__data__ = new MapCache(pairs);
  ***REMOVED***
  cache.set(key, value);
  return this;
***REMOVED***

// Add methods to `Stack`.
Stack.prototype.clear = stackClear;
Stack.prototype['delete'] = stackDelete;
Stack.prototype.get = stackGet;
Stack.prototype.has = stackHas;
Stack.prototype.set = stackSet;

/**
 * Creates an array of the enumerable property names of the array-like `value`.
 *
 * @private
 * @param {****REMOVED*** value The value to query.
 * @param {boolean***REMOVED*** inherited Specify returning inherited property names.
 * @returns {Array***REMOVED*** Returns the array of property names.
 */
function arrayLikeKeys(value, inherited) {
  // Safari 8.1 makes `arguments.callee` enumerable in strict mode.
  // Safari 9 makes `arguments.length` enumerable in strict mode.
  var result = (isArray(value) || isArguments(value))
    ? baseTimes(value.length, String)
    : [];

  var length = result.length,
      skipIndexes = !!length;

  for (var key in value) {
    if ((inherited || hasOwnProperty.call(value, key)) &&
        !(skipIndexes && (key == 'length' || isIndex(key, length)))) {
      result.push(key);
    ***REMOVED***
  ***REMOVED***
  return result;
***REMOVED***

/**
 * Gets the index at which the `key` is found in `array` of key-value pairs.
 *
 * @private
 * @param {Array***REMOVED*** array The array to inspect.
 * @param {****REMOVED*** key The key to search for.
 * @returns {number***REMOVED*** Returns the index of the matched value, else `-1`.
 */
function assocIndexOf(array, key) {
  var length = array.length;
  while (length--) {
    if (eq(array[length][0], key)) {
      return length;
    ***REMOVED***
  ***REMOVED***
  return -1;
***REMOVED***

/**
 * The base implementation of `_.forEach` without support for iteratee shorthands.
 *
 * @private
 * @param {Array|Object***REMOVED*** collection The collection to iterate over.
 * @param {Function***REMOVED*** iteratee The function invoked per iteration.
 * @returns {Array|Object***REMOVED*** Returns `collection`.
 */
var baseEach = createBaseEach(baseForOwn);

/**
 * The base implementation of `_.flatten` with support for restricting flattening.
 *
 * @private
 * @param {Array***REMOVED*** array The array to flatten.
 * @param {number***REMOVED*** depth The maximum recursion depth.
 * @param {boolean***REMOVED*** [predicate=isFlattenable] The function invoked per iteration.
 * @param {boolean***REMOVED*** [isStrict] Restrict to values that pass `predicate` checks.
 * @param {Array***REMOVED*** [result=[]] The initial result value.
 * @returns {Array***REMOVED*** Returns the new flattened array.
 */
function baseFlatten(array, depth, predicate, isStrict, result) {
  var index = -1,
      length = array.length;

  predicate || (predicate = isFlattenable);
  result || (result = []);

  while (++index < length) {
    var value = array[index];
    if (depth > 0 && predicate(value)) {
      if (depth > 1) {
        // Recursively flatten arrays (susceptible to call stack limits).
        baseFlatten(value, depth - 1, predicate, isStrict, result);
      ***REMOVED*** else {
        arrayPush(result, value);
      ***REMOVED***
    ***REMOVED*** else if (!isStrict) {
      result[result.length] = value;
    ***REMOVED***
  ***REMOVED***
  return result;
***REMOVED***

/**
 * The base implementation of `baseForOwn` which iterates over `object`
 * properties returned by `keysFunc` and invokes `iteratee` for each property.
 * Iteratee functions may exit iteration early by explicitly returning `false`.
 *
 * @private
 * @param {Object***REMOVED*** object The object to iterate over.
 * @param {Function***REMOVED*** iteratee The function invoked per iteration.
 * @param {Function***REMOVED*** keysFunc The function to get the keys of `object`.
 * @returns {Object***REMOVED*** Returns `object`.
 */
var baseFor = createBaseFor();

/**
 * The base implementation of `_.forOwn` without support for iteratee shorthands.
 *
 * @private
 * @param {Object***REMOVED*** object The object to iterate over.
 * @param {Function***REMOVED*** iteratee The function invoked per iteration.
 * @returns {Object***REMOVED*** Returns `object`.
 */
function baseForOwn(object, iteratee) {
  return object && baseFor(object, iteratee, keys);
***REMOVED***

/**
 * The base implementation of `_.get` without support for default values.
 *
 * @private
 * @param {Object***REMOVED*** object The object to query.
 * @param {Array|string***REMOVED*** path The path of the property to get.
 * @returns {****REMOVED*** Returns the resolved value.
 */
function baseGet(object, path) {
  path = isKey(path, object) ? [path] : castPath(path);

  var index = 0,
      length = path.length;

  while (object != null && index < length) {
    object = object[toKey(path[index++])];
  ***REMOVED***
  return (index && index == length) ? object : undefined;
***REMOVED***

/**
 * The base implementation of `getTag`.
 *
 * @private
 * @param {****REMOVED*** value The value to query.
 * @returns {string***REMOVED*** Returns the `toStringTag`.
 */
function baseGetTag(value) {
  return objectToString.call(value);
***REMOVED***

/**
 * The base implementation of `_.hasIn` without support for deep paths.
 *
 * @private
 * @param {Object***REMOVED*** [object] The object to query.
 * @param {Array|string***REMOVED*** key The key to check.
 * @returns {boolean***REMOVED*** Returns `true` if `key` exists, else `false`.
 */
function baseHasIn(object, key) {
  return object != null && key in Object(object);
***REMOVED***

/**
 * The base implementation of `_.isEqual` which supports partial comparisons
 * and tracks traversed objects.
 *
 * @private
 * @param {****REMOVED*** value The value to compare.
 * @param {****REMOVED*** other The other value to compare.
 * @param {Function***REMOVED*** [customizer] The function to customize comparisons.
 * @param {boolean***REMOVED*** [bitmask] The bitmask of comparison flags.
 *  The bitmask may be composed of the following flags:
 *     1 - Unordered comparison
 *     2 - Partial comparison
 * @param {Object***REMOVED*** [stack] Tracks traversed `value` and `other` objects.
 * @returns {boolean***REMOVED*** Returns `true` if the values are equivalent, else `false`.
 */
function baseIsEqual(value, other, customizer, bitmask, stack) {
  if (value === other) {
    return true;
  ***REMOVED***
  if (value == null || other == null || (!isObject(value) && !isObjectLike(other))) {
    return value !== value && other !== other;
  ***REMOVED***
  return baseIsEqualDeep(value, other, baseIsEqual, customizer, bitmask, stack);
***REMOVED***

/**
 * A specialized version of `baseIsEqual` for arrays and objects which performs
 * deep comparisons and tracks traversed objects enabling objects with circular
 * references to be compared.
 *
 * @private
 * @param {Object***REMOVED*** object The object to compare.
 * @param {Object***REMOVED*** other The other object to compare.
 * @param {Function***REMOVED*** equalFunc The function to determine equivalents of values.
 * @param {Function***REMOVED*** [customizer] The function to customize comparisons.
 * @param {number***REMOVED*** [bitmask] The bitmask of comparison flags. See `baseIsEqual`
 *  for more details.
 * @param {Object***REMOVED*** [stack] Tracks traversed `object` and `other` objects.
 * @returns {boolean***REMOVED*** Returns `true` if the objects are equivalent, else `false`.
 */
function baseIsEqualDeep(object, other, equalFunc, customizer, bitmask, stack) {
  var objIsArr = isArray(object),
      othIsArr = isArray(other),
      objTag = arrayTag,
      othTag = arrayTag;

  if (!objIsArr) {
    objTag = getTag(object);
    objTag = objTag == argsTag ? objectTag : objTag;
  ***REMOVED***
  if (!othIsArr) {
    othTag = getTag(other);
    othTag = othTag == argsTag ? objectTag : othTag;
  ***REMOVED***
  var objIsObj = objTag == objectTag && !isHostObject(object),
      othIsObj = othTag == objectTag && !isHostObject(other),
      isSameTag = objTag == othTag;

  if (isSameTag && !objIsObj) {
    stack || (stack = new Stack);
    return (objIsArr || isTypedArray(object))
      ? equalArrays(object, other, equalFunc, customizer, bitmask, stack)
      : equalByTag(object, other, objTag, equalFunc, customizer, bitmask, stack);
  ***REMOVED***
  if (!(bitmask & PARTIAL_COMPARE_FLAG)) {
    var objIsWrapped = objIsObj && hasOwnProperty.call(object, '__wrapped__'),
        othIsWrapped = othIsObj && hasOwnProperty.call(other, '__wrapped__');

    if (objIsWrapped || othIsWrapped) {
      var objUnwrapped = objIsWrapped ? object.value() : object,
          othUnwrapped = othIsWrapped ? other.value() : other;

      stack || (stack = new Stack);
      return equalFunc(objUnwrapped, othUnwrapped, customizer, bitmask, stack);
    ***REMOVED***
  ***REMOVED***
  if (!isSameTag) {
    return false;
  ***REMOVED***
  stack || (stack = new Stack);
  return equalObjects(object, other, equalFunc, customizer, bitmask, stack);
***REMOVED***

/**
 * The base implementation of `_.isMatch` without support for iteratee shorthands.
 *
 * @private
 * @param {Object***REMOVED*** object The object to inspect.
 * @param {Object***REMOVED*** source The object of property values to match.
 * @param {Array***REMOVED*** matchData The property names, values, and compare flags to match.
 * @param {Function***REMOVED*** [customizer] The function to customize comparisons.
 * @returns {boolean***REMOVED*** Returns `true` if `object` is a match, else `false`.
 */
function baseIsMatch(object, source, matchData, customizer) {
  var index = matchData.length,
      length = index,
      noCustomizer = !customizer;

  if (object == null) {
    return !length;
  ***REMOVED***
  object = Object(object);
  while (index--) {
    var data = matchData[index];
    if ((noCustomizer && data[2])
          ? data[1] !== object[data[0]]
          : !(data[0] in object)
        ) {
      return false;
    ***REMOVED***
  ***REMOVED***
  while (++index < length) {
    data = matchData[index];
    var key = data[0],
        objValue = object[key],
        srcValue = data[1];

    if (noCustomizer && data[2]) {
      if (objValue === undefined && !(key in object)) {
        return false;
      ***REMOVED***
    ***REMOVED*** else {
      var stack = new Stack;
      if (customizer) {
        var result = customizer(objValue, srcValue, key, object, source, stack);
      ***REMOVED***
      if (!(result === undefined
            ? baseIsEqual(srcValue, objValue, customizer, UNORDERED_COMPARE_FLAG | PARTIAL_COMPARE_FLAG, stack)
            : result
          )) {
        return false;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  return true;
***REMOVED***

/**
 * The base implementation of `_.isNative` without bad shim checks.
 *
 * @private
 * @param {****REMOVED*** value The value to check.
 * @returns {boolean***REMOVED*** Returns `true` if `value` is a native function,
 *  else `false`.
 */
function baseIsNative(value) {
  if (!isObject(value) || isMasked(value)) {
    return false;
  ***REMOVED***
  var pattern = (isFunction(value) || isHostObject(value)) ? reIsNative : reIsHostCtor;
  return pattern.test(toSource(value));
***REMOVED***

/**
 * The base implementation of `_.isTypedArray` without Node.js optimizations.
 *
 * @private
 * @param {****REMOVED*** value The value to check.
 * @returns {boolean***REMOVED*** Returns `true` if `value` is a typed array, else `false`.
 */
function baseIsTypedArray(value) {
  return isObjectLike(value) &&
    isLength(value.length) && !!typedArrayTags[objectToString.call(value)];
***REMOVED***

/**
 * The base implementation of `_.iteratee`.
 *
 * @private
 * @param {****REMOVED*** [value=_.identity] The value to convert to an iteratee.
 * @returns {Function***REMOVED*** Returns the iteratee.
 */
function baseIteratee(value) {
  // Don't store the `typeof` result in a variable to avoid a JIT bug in Safari 9.
  // See https://bugs.webkit.org/show_bug.cgi?id=156034 for more details.
  if (typeof value == 'function') {
    return value;
  ***REMOVED***
  if (value == null) {
    return identity;
  ***REMOVED***
  if (typeof value == 'object') {
    return isArray(value)
      ? baseMatchesProperty(value[0], value[1])
      : baseMatches(value);
  ***REMOVED***
  return property(value);
***REMOVED***

/**
 * The base implementation of `_.keys` which doesn't treat sparse arrays as dense.
 *
 * @private
 * @param {Object***REMOVED*** object The object to query.
 * @returns {Array***REMOVED*** Returns the array of property names.
 */
function baseKeys(object) {
  if (!isPrototype(object)) {
    return nativeKeys(object);
  ***REMOVED***
  var result = [];
  for (var key in Object(object)) {
    if (hasOwnProperty.call(object, key) && key != 'constructor') {
      result.push(key);
    ***REMOVED***
  ***REMOVED***
  return result;
***REMOVED***

/**
 * The base implementation of `_.map` without support for iteratee shorthands.
 *
 * @private
 * @param {Array|Object***REMOVED*** collection The collection to iterate over.
 * @param {Function***REMOVED*** iteratee The function invoked per iteration.
 * @returns {Array***REMOVED*** Returns the new mapped array.
 */
function baseMap(collection, iteratee) {
  var index = -1,
      result = isArrayLike(collection) ? Array(collection.length) : [];

  baseEach(collection, function(value, key, collection) {
    result[++index] = iteratee(value, key, collection);
  ***REMOVED***);
  return result;
***REMOVED***

/**
 * The base implementation of `_.matches` which doesn't clone `source`.
 *
 * @private
 * @param {Object***REMOVED*** source The object of property values to match.
 * @returns {Function***REMOVED*** Returns the new spec function.
 */
function baseMatches(source) {
  var matchData = getMatchData(source);
  if (matchData.length == 1 && matchData[0][2]) {
    return matchesStrictComparable(matchData[0][0], matchData[0][1]);
  ***REMOVED***
  return function(object) {
    return object === source || baseIsMatch(object, source, matchData);
  ***REMOVED***;
***REMOVED***

/**
 * The base implementation of `_.matchesProperty` which doesn't clone `srcValue`.
 *
 * @private
 * @param {string***REMOVED*** path The path of the property to get.
 * @param {****REMOVED*** srcValue The value to match.
 * @returns {Function***REMOVED*** Returns the new spec function.
 */
function baseMatchesProperty(path, srcValue) {
  if (isKey(path) && isStrictComparable(srcValue)) {
    return matchesStrictComparable(toKey(path), srcValue);
  ***REMOVED***
  return function(object) {
    var objValue = get(object, path);
    return (objValue === undefined && objValue === srcValue)
      ? hasIn(object, path)
      : baseIsEqual(srcValue, objValue, undefined, UNORDERED_COMPARE_FLAG | PARTIAL_COMPARE_FLAG);
  ***REMOVED***;
***REMOVED***

/**
 * The base implementation of `_.orderBy` without param guards.
 *
 * @private
 * @param {Array|Object***REMOVED*** collection The collection to iterate over.
 * @param {Function[]|Object[]|string[]***REMOVED*** iteratees The iteratees to sort by.
 * @param {string[]***REMOVED*** orders The sort orders of `iteratees`.
 * @returns {Array***REMOVED*** Returns the new sorted array.
 */
function baseOrderBy(collection, iteratees, orders) {
  var index = -1;
  iteratees = arrayMap(iteratees.length ? iteratees : [identity], baseUnary(baseIteratee));

  var result = baseMap(collection, function(value, key, collection) {
    var criteria = arrayMap(iteratees, function(iteratee) {
      return iteratee(value);
    ***REMOVED***);
    return { 'criteria': criteria, 'index': ++index, 'value': value ***REMOVED***;
  ***REMOVED***);

  return baseSortBy(result, function(object, other) {
    return compareMultiple(object, other, orders);
  ***REMOVED***);
***REMOVED***

/**
 * A specialized version of `baseProperty` which supports deep paths.
 *
 * @private
 * @param {Array|string***REMOVED*** path The path of the property to get.
 * @returns {Function***REMOVED*** Returns the new accessor function.
 */
function basePropertyDeep(path) {
  return function(object) {
    return baseGet(object, path);
  ***REMOVED***;
***REMOVED***

/**
 * The base implementation of `_.rest` which doesn't validate or coerce arguments.
 *
 * @private
 * @param {Function***REMOVED*** func The function to apply a rest parameter to.
 * @param {number***REMOVED*** [start=func.length-1] The start position of the rest parameter.
 * @returns {Function***REMOVED*** Returns the new function.
 */
function baseRest(func, start) {
  start = nativeMax(start === undefined ? (func.length - 1) : start, 0);
  return function() {
    var args = arguments,
        index = -1,
        length = nativeMax(args.length - start, 0),
        array = Array(length);

    while (++index < length) {
      array[index] = args[start + index];
    ***REMOVED***
    index = -1;
    var otherArgs = Array(start + 1);
    while (++index < start) {
      otherArgs[index] = args[index];
    ***REMOVED***
    otherArgs[start] = array;
    return apply(func, this, otherArgs);
  ***REMOVED***;
***REMOVED***

/**
 * The base implementation of `_.toString` which doesn't convert nullish
 * values to empty strings.
 *
 * @private
 * @param {****REMOVED*** value The value to process.
 * @returns {string***REMOVED*** Returns the string.
 */
function baseToString(value) {
  // Exit early for strings to avoid a performance hit in some environments.
  if (typeof value == 'string') {
    return value;
  ***REMOVED***
  if (isSymbol(value)) {
    return symbolToString ? symbolToString.call(value) : '';
  ***REMOVED***
  var result = (value + '');
  return (result == '0' && (1 / value) == -INFINITY) ? '-0' : result;
***REMOVED***

/**
 * Casts `value` to a path array if it's not one.
 *
 * @private
 * @param {****REMOVED*** value The value to inspect.
 * @returns {Array***REMOVED*** Returns the cast property path array.
 */
function castPath(value) {
  return isArray(value) ? value : stringToPath(value);
***REMOVED***

/**
 * Compares values to sort them in ascending order.
 *
 * @private
 * @param {****REMOVED*** value The value to compare.
 * @param {****REMOVED*** other The other value to compare.
 * @returns {number***REMOVED*** Returns the sort order indicator for `value`.
 */
function compareAscending(value, other) {
  if (value !== other) {
    var valIsDefined = value !== undefined,
        valIsNull = value === null,
        valIsReflexive = value === value,
        valIsSymbol = isSymbol(value);

    var othIsDefined = other !== undefined,
        othIsNull = other === null,
        othIsReflexive = other === other,
        othIsSymbol = isSymbol(other);

    if ((!othIsNull && !othIsSymbol && !valIsSymbol && value > other) ||
        (valIsSymbol && othIsDefined && othIsReflexive && !othIsNull && !othIsSymbol) ||
        (valIsNull && othIsDefined && othIsReflexive) ||
        (!valIsDefined && othIsReflexive) ||
        !valIsReflexive) {
      return 1;
    ***REMOVED***
    if ((!valIsNull && !valIsSymbol && !othIsSymbol && value < other) ||
        (othIsSymbol && valIsDefined && valIsReflexive && !valIsNull && !valIsSymbol) ||
        (othIsNull && valIsDefined && valIsReflexive) ||
        (!othIsDefined && valIsReflexive) ||
        !othIsReflexive) {
      return -1;
    ***REMOVED***
  ***REMOVED***
  return 0;
***REMOVED***

/**
 * Used by `_.orderBy` to compare multiple properties of a value to another
 * and stable sort them.
 *
 * If `orders` is unspecified, all values are sorted in ascending order. Otherwise,
 * specify an order of "desc" for descending or "asc" for ascending sort order
 * of corresponding values.
 *
 * @private
 * @param {Object***REMOVED*** object The object to compare.
 * @param {Object***REMOVED*** other The other object to compare.
 * @param {boolean[]|string[]***REMOVED*** orders The order to sort by for each property.
 * @returns {number***REMOVED*** Returns the sort order indicator for `object`.
 */
function compareMultiple(object, other, orders) {
  var index = -1,
      objCriteria = object.criteria,
      othCriteria = other.criteria,
      length = objCriteria.length,
      ordersLength = orders.length;

  while (++index < length) {
    var result = compareAscending(objCriteria[index], othCriteria[index]);
    if (result) {
      if (index >= ordersLength) {
        return result;
      ***REMOVED***
      var order = orders[index];
      return result * (order == 'desc' ? -1 : 1);
    ***REMOVED***
  ***REMOVED***
  // Fixes an `Array#sort` bug in the JS engine embedded in Adobe applications
  // that causes it, under certain circumstances, to provide the same value for
  // `object` and `other`. See https://github.com/jashkenas/underscore/pull/1247
  // for more details.
  //
  // This also ensures a stable sort in V8 and other engines.
  // See https://bugs.chromium.org/p/v8/issues/detail?id=90 for more details.
  return object.index - other.index;
***REMOVED***

/**
 * Creates a `baseEach` or `baseEachRight` function.
 *
 * @private
 * @param {Function***REMOVED*** eachFunc The function to iterate over a collection.
 * @param {boolean***REMOVED*** [fromRight] Specify iterating from right to left.
 * @returns {Function***REMOVED*** Returns the new base function.
 */
function createBaseEach(eachFunc, fromRight) {
  return function(collection, iteratee) {
    if (collection == null) {
      return collection;
    ***REMOVED***
    if (!isArrayLike(collection)) {
      return eachFunc(collection, iteratee);
    ***REMOVED***
    var length = collection.length,
        index = fromRight ? length : -1,
        iterable = Object(collection);

    while ((fromRight ? index-- : ++index < length)) {
      if (iteratee(iterable[index], index, iterable) === false) {
        break;
      ***REMOVED***
    ***REMOVED***
    return collection;
  ***REMOVED***;
***REMOVED***

/**
 * Creates a base function for methods like `_.forIn` and `_.forOwn`.
 *
 * @private
 * @param {boolean***REMOVED*** [fromRight] Specify iterating from right to left.
 * @returns {Function***REMOVED*** Returns the new base function.
 */
function createBaseFor(fromRight) {
  return function(object, iteratee, keysFunc) {
    var index = -1,
        iterable = Object(object),
        props = keysFunc(object),
        length = props.length;

    while (length--) {
      var key = props[fromRight ? length : ++index];
      if (iteratee(iterable[key], key, iterable) === false) {
        break;
      ***REMOVED***
    ***REMOVED***
    return object;
  ***REMOVED***;
***REMOVED***

/**
 * A specialized version of `baseIsEqualDeep` for arrays with support for
 * partial deep comparisons.
 *
 * @private
 * @param {Array***REMOVED*** array The array to compare.
 * @param {Array***REMOVED*** other The other array to compare.
 * @param {Function***REMOVED*** equalFunc The function to determine equivalents of values.
 * @param {Function***REMOVED*** customizer The function to customize comparisons.
 * @param {number***REMOVED*** bitmask The bitmask of comparison flags. See `baseIsEqual`
 *  for more details.
 * @param {Object***REMOVED*** stack Tracks traversed `array` and `other` objects.
 * @returns {boolean***REMOVED*** Returns `true` if the arrays are equivalent, else `false`.
 */
function equalArrays(array, other, equalFunc, customizer, bitmask, stack) {
  var isPartial = bitmask & PARTIAL_COMPARE_FLAG,
      arrLength = array.length,
      othLength = other.length;

  if (arrLength != othLength && !(isPartial && othLength > arrLength)) {
    return false;
  ***REMOVED***
  // Assume cyclic values are equal.
  var stacked = stack.get(array);
  if (stacked && stack.get(other)) {
    return stacked == other;
  ***REMOVED***
  var index = -1,
      result = true,
      seen = (bitmask & UNORDERED_COMPARE_FLAG) ? new SetCache : undefined;

  stack.set(array, other);
  stack.set(other, array);

  // Ignore non-index properties.
  while (++index < arrLength) {
    var arrValue = array[index],
        othValue = other[index];

    if (customizer) {
      var compared = isPartial
        ? customizer(othValue, arrValue, index, other, array, stack)
        : customizer(arrValue, othValue, index, array, other, stack);
    ***REMOVED***
    if (compared !== undefined) {
      if (compared) {
        continue;
      ***REMOVED***
      result = false;
      break;
    ***REMOVED***
    // Recursively compare arrays (susceptible to call stack limits).
    if (seen) {
      if (!arraySome(other, function(othValue, othIndex) {
            if (!seen.has(othIndex) &&
                (arrValue === othValue || equalFunc(arrValue, othValue, customizer, bitmask, stack))) {
              return seen.add(othIndex);
            ***REMOVED***
          ***REMOVED***)) {
        result = false;
        break;
      ***REMOVED***
    ***REMOVED*** else if (!(
          arrValue === othValue ||
            equalFunc(arrValue, othValue, customizer, bitmask, stack)
        )) {
      result = false;
      break;
    ***REMOVED***
  ***REMOVED***
  stack['delete'](array);
  stack['delete'](other);
  return result;
***REMOVED***

/**
 * A specialized version of `baseIsEqualDeep` for comparing objects of
 * the same `toStringTag`.
 *
 * **Note:** This function only supports comparing values with tags of
 * `Boolean`, `Date`, `Error`, `Number`, `RegExp`, or `String`.
 *
 * @private
 * @param {Object***REMOVED*** object The object to compare.
 * @param {Object***REMOVED*** other The other object to compare.
 * @param {string***REMOVED*** tag The `toStringTag` of the objects to compare.
 * @param {Function***REMOVED*** equalFunc The function to determine equivalents of values.
 * @param {Function***REMOVED*** customizer The function to customize comparisons.
 * @param {number***REMOVED*** bitmask The bitmask of comparison flags. See `baseIsEqual`
 *  for more details.
 * @param {Object***REMOVED*** stack Tracks traversed `object` and `other` objects.
 * @returns {boolean***REMOVED*** Returns `true` if the objects are equivalent, else `false`.
 */
function equalByTag(object, other, tag, equalFunc, customizer, bitmask, stack) {
  switch (tag) {
    case dataViewTag:
      if ((object.byteLength != other.byteLength) ||
          (object.byteOffset != other.byteOffset)) {
        return false;
      ***REMOVED***
      object = object.buffer;
      other = other.buffer;

    case arrayBufferTag:
      if ((object.byteLength != other.byteLength) ||
          !equalFunc(new Uint8Array(object), new Uint8Array(other))) {
        return false;
      ***REMOVED***
      return true;

    case boolTag:
    case dateTag:
    case numberTag:
      // Coerce booleans to `1` or `0` and dates to milliseconds.
      // Invalid dates are coerced to `NaN`.
      return eq(+object, +other);

    case errorTag:
      return object.name == other.name && object.message == other.message;

    case regexpTag:
    case stringTag:
      // Coerce regexes to strings and treat strings, primitives and objects,
      // as equal. See http://www.ecma-international.org/ecma-262/7.0/#sec-regexp.prototype.tostring
      // for more details.
      return object == (other + '');

    case mapTag:
      var convert = mapToArray;

    case setTag:
      var isPartial = bitmask & PARTIAL_COMPARE_FLAG;
      convert || (convert = setToArray);

      if (object.size != other.size && !isPartial) {
        return false;
      ***REMOVED***
      // Assume cyclic values are equal.
      var stacked = stack.get(object);
      if (stacked) {
        return stacked == other;
      ***REMOVED***
      bitmask |= UNORDERED_COMPARE_FLAG;

      // Recursively compare objects (susceptible to call stack limits).
      stack.set(object, other);
      var result = equalArrays(convert(object), convert(other), equalFunc, customizer, bitmask, stack);
      stack['delete'](object);
      return result;

    case symbolTag:
      if (symbolValueOf) {
        return symbolValueOf.call(object) == symbolValueOf.call(other);
      ***REMOVED***
  ***REMOVED***
  return false;
***REMOVED***

/**
 * A specialized version of `baseIsEqualDeep` for objects with support for
 * partial deep comparisons.
 *
 * @private
 * @param {Object***REMOVED*** object The object to compare.
 * @param {Object***REMOVED*** other The other object to compare.
 * @param {Function***REMOVED*** equalFunc The function to determine equivalents of values.
 * @param {Function***REMOVED*** customizer The function to customize comparisons.
 * @param {number***REMOVED*** bitmask The bitmask of comparison flags. See `baseIsEqual`
 *  for more details.
 * @param {Object***REMOVED*** stack Tracks traversed `object` and `other` objects.
 * @returns {boolean***REMOVED*** Returns `true` if the objects are equivalent, else `false`.
 */
function equalObjects(object, other, equalFunc, customizer, bitmask, stack) {
  var isPartial = bitmask & PARTIAL_COMPARE_FLAG,
      objProps = keys(object),
      objLength = objProps.length,
      othProps = keys(other),
      othLength = othProps.length;

  if (objLength != othLength && !isPartial) {
    return false;
  ***REMOVED***
  var index = objLength;
  while (index--) {
    var key = objProps[index];
    if (!(isPartial ? key in other : hasOwnProperty.call(other, key))) {
      return false;
    ***REMOVED***
  ***REMOVED***
  // Assume cyclic values are equal.
  var stacked = stack.get(object);
  if (stacked && stack.get(other)) {
    return stacked == other;
  ***REMOVED***
  var result = true;
  stack.set(object, other);
  stack.set(other, object);

  var skipCtor = isPartial;
  while (++index < objLength) {
    key = objProps[index];
    var objValue = object[key],
        othValue = other[key];

    if (customizer) {
      var compared = isPartial
        ? customizer(othValue, objValue, key, other, object, stack)
        : customizer(objValue, othValue, key, object, other, stack);
    ***REMOVED***
    // Recursively compare objects (susceptible to call stack limits).
    if (!(compared === undefined
          ? (objValue === othValue || equalFunc(objValue, othValue, customizer, bitmask, stack))
          : compared
        )) {
      result = false;
      break;
    ***REMOVED***
    skipCtor || (skipCtor = key == 'constructor');
  ***REMOVED***
  if (result && !skipCtor) {
    var objCtor = object.constructor,
        othCtor = other.constructor;

    // Non `Object` object instances with different constructors are not equal.
    if (objCtor != othCtor &&
        ('constructor' in object && 'constructor' in other) &&
        !(typeof objCtor == 'function' && objCtor instanceof objCtor &&
          typeof othCtor == 'function' && othCtor instanceof othCtor)) {
      result = false;
    ***REMOVED***
  ***REMOVED***
  stack['delete'](object);
  stack['delete'](other);
  return result;
***REMOVED***

/**
 * Gets the data for `map`.
 *
 * @private
 * @param {Object***REMOVED*** map The map to query.
 * @param {string***REMOVED*** key The reference key.
 * @returns {****REMOVED*** Returns the map data.
 */
function getMapData(map, key) {
  var data = map.__data__;
  return isKeyable(key)
    ? data[typeof key == 'string' ? 'string' : 'hash']
    : data.map;
***REMOVED***

/**
 * Gets the property names, values, and compare flags of `object`.
 *
 * @private
 * @param {Object***REMOVED*** object The object to query.
 * @returns {Array***REMOVED*** Returns the match data of `object`.
 */
function getMatchData(object) {
  var result = keys(object),
      length = result.length;

  while (length--) {
    var key = result[length],
        value = object[key];

    result[length] = [key, value, isStrictComparable(value)];
  ***REMOVED***
  return result;
***REMOVED***

/**
 * Gets the native function at `key` of `object`.
 *
 * @private
 * @param {Object***REMOVED*** object The object to query.
 * @param {string***REMOVED*** key The key of the method to get.
 * @returns {****REMOVED*** Returns the function if it's native, else `undefined`.
 */
function getNative(object, key) {
  var value = getValue(object, key);
  return baseIsNative(value) ? value : undefined;
***REMOVED***

/**
 * Gets the `toStringTag` of `value`.
 *
 * @private
 * @param {****REMOVED*** value The value to query.
 * @returns {string***REMOVED*** Returns the `toStringTag`.
 */
var getTag = baseGetTag;

// Fallback for data views, maps, sets, and weak maps in IE 11,
// for data views in Edge < 14, and promises in Node.js.
if ((DataView && getTag(new DataView(new ArrayBuffer(1))) != dataViewTag) ||
    (Map && getTag(new Map) != mapTag) ||
    (Promise && getTag(Promise.resolve()) != promiseTag) ||
    (Set && getTag(new Set) != setTag) ||
    (WeakMap && getTag(new WeakMap) != weakMapTag)) {
  getTag = function(value) {
    var result = objectToString.call(value),
        Ctor = result == objectTag ? value.constructor : undefined,
        ctorString = Ctor ? toSource(Ctor) : undefined;

    if (ctorString) {
      switch (ctorString) {
        case dataViewCtorString: return dataViewTag;
        case mapCtorString: return mapTag;
        case promiseCtorString: return promiseTag;
        case setCtorString: return setTag;
        case weakMapCtorString: return weakMapTag;
      ***REMOVED***
    ***REMOVED***
    return result;
  ***REMOVED***;
***REMOVED***

/**
 * Checks if `path` exists on `object`.
 *
 * @private
 * @param {Object***REMOVED*** object The object to query.
 * @param {Array|string***REMOVED*** path The path to check.
 * @param {Function***REMOVED*** hasFunc The function to check properties.
 * @returns {boolean***REMOVED*** Returns `true` if `path` exists, else `false`.
 */
function hasPath(object, path, hasFunc) {
  path = isKey(path, object) ? [path] : castPath(path);

  var result,
      index = -1,
      length = path.length;

  while (++index < length) {
    var key = toKey(path[index]);
    if (!(result = object != null && hasFunc(object, key))) {
      break;
    ***REMOVED***
    object = object[key];
  ***REMOVED***
  if (result) {
    return result;
  ***REMOVED***
  var length = object ? object.length : 0;
  return !!length && isLength(length) && isIndex(key, length) &&
    (isArray(object) || isArguments(object));
***REMOVED***

/**
 * Checks if `value` is a flattenable `arguments` object or array.
 *
 * @private
 * @param {****REMOVED*** value The value to check.
 * @returns {boolean***REMOVED*** Returns `true` if `value` is flattenable, else `false`.
 */
function isFlattenable(value) {
  return isArray(value) || isArguments(value) ||
    !!(spreadableSymbol && value && value[spreadableSymbol]);
***REMOVED***

/**
 * Checks if `value` is a valid array-like index.
 *
 * @private
 * @param {****REMOVED*** value The value to check.
 * @param {number***REMOVED*** [length=MAX_SAFE_INTEGER] The upper bounds of a valid index.
 * @returns {boolean***REMOVED*** Returns `true` if `value` is a valid index, else `false`.
 */
function isIndex(value, length) {
  length = length == null ? MAX_SAFE_INTEGER : length;
  return !!length &&
    (typeof value == 'number' || reIsUint.test(value)) &&
    (value > -1 && value % 1 == 0 && value < length);
***REMOVED***

/**
 * Checks if the given arguments are from an iteratee call.
 *
 * @private
 * @param {****REMOVED*** value The potential iteratee value argument.
 * @param {****REMOVED*** index The potential iteratee index or key argument.
 * @param {****REMOVED*** object The potential iteratee object argument.
 * @returns {boolean***REMOVED*** Returns `true` if the arguments are from an iteratee call,
 *  else `false`.
 */
function isIterateeCall(value, index, object) {
  if (!isObject(object)) {
    return false;
  ***REMOVED***
  var type = typeof index;
  if (type == 'number'
        ? (isArrayLike(object) && isIndex(index, object.length))
        : (type == 'string' && index in object)
      ) {
    return eq(object[index], value);
  ***REMOVED***
  return false;
***REMOVED***

/**
 * Checks if `value` is a property name and not a property path.
 *
 * @private
 * @param {****REMOVED*** value The value to check.
 * @param {Object***REMOVED*** [object] The object to query keys on.
 * @returns {boolean***REMOVED*** Returns `true` if `value` is a property name, else `false`.
 */
function isKey(value, object) {
  if (isArray(value)) {
    return false;
  ***REMOVED***
  var type = typeof value;
  if (type == 'number' || type == 'symbol' || type == 'boolean' ||
      value == null || isSymbol(value)) {
    return true;
  ***REMOVED***
  return reIsPlainProp.test(value) || !reIsDeepProp.test(value) ||
    (object != null && value in Object(object));
***REMOVED***

/**
 * Checks if `value` is suitable for use as unique object key.
 *
 * @private
 * @param {****REMOVED*** value The value to check.
 * @returns {boolean***REMOVED*** Returns `true` if `value` is suitable, else `false`.
 */
function isKeyable(value) {
  var type = typeof value;
  return (type == 'string' || type == 'number' || type == 'symbol' || type == 'boolean')
    ? (value !== '__proto__')
    : (value === null);
***REMOVED***

/**
 * Checks if `func` has its source masked.
 *
 * @private
 * @param {Function***REMOVED*** func The function to check.
 * @returns {boolean***REMOVED*** Returns `true` if `func` is masked, else `false`.
 */
function isMasked(func) {
  return !!maskSrcKey && (maskSrcKey in func);
***REMOVED***

/**
 * Checks if `value` is likely a prototype object.
 *
 * @private
 * @param {****REMOVED*** value The value to check.
 * @returns {boolean***REMOVED*** Returns `true` if `value` is a prototype, else `false`.
 */
function isPrototype(value) {
  var Ctor = value && value.constructor,
      proto = (typeof Ctor == 'function' && Ctor.prototype) || objectProto;

  return value === proto;
***REMOVED***

/**
 * Checks if `value` is suitable for strict equality comparisons, i.e. `===`.
 *
 * @private
 * @param {****REMOVED*** value The value to check.
 * @returns {boolean***REMOVED*** Returns `true` if `value` if suitable for strict
 *  equality comparisons, else `false`.
 */
function isStrictComparable(value) {
  return value === value && !isObject(value);
***REMOVED***

/**
 * A specialized version of `matchesProperty` for source values suitable
 * for strict equality comparisons, i.e. `===`.
 *
 * @private
 * @param {string***REMOVED*** key The key of the property to get.
 * @param {****REMOVED*** srcValue The value to match.
 * @returns {Function***REMOVED*** Returns the new spec function.
 */
function matchesStrictComparable(key, srcValue) {
  return function(object) {
    if (object == null) {
      return false;
    ***REMOVED***
    return object[key] === srcValue &&
      (srcValue !== undefined || (key in Object(object)));
  ***REMOVED***;
***REMOVED***

/**
 * Converts `string` to a property path array.
 *
 * @private
 * @param {string***REMOVED*** string The string to convert.
 * @returns {Array***REMOVED*** Returns the property path array.
 */
var stringToPath = memoize(function(string) {
  string = toString(string);

  var result = [];
  if (reLeadingDot.test(string)) {
    result.push('');
  ***REMOVED***
  string.replace(rePropName, function(match, number, quote, string) {
    result.push(quote ? string.replace(reEscapeChar, '$1') : (number || match));
  ***REMOVED***);
  return result;
***REMOVED***);

/**
 * Converts `value` to a string key if it's not a string or symbol.
 *
 * @private
 * @param {****REMOVED*** value The value to inspect.
 * @returns {string|symbol***REMOVED*** Returns the key.
 */
function toKey(value) {
  if (typeof value == 'string' || isSymbol(value)) {
    return value;
  ***REMOVED***
  var result = (value + '');
  return (result == '0' && (1 / value) == -INFINITY) ? '-0' : result;
***REMOVED***

/**
 * Converts `func` to its source code.
 *
 * @private
 * @param {Function***REMOVED*** func The function to process.
 * @returns {string***REMOVED*** Returns the source code.
 */
function toSource(func) {
  if (func != null) {
    try {
      return funcToString.call(func);
    ***REMOVED*** catch (e) {***REMOVED***
    try {
      return (func + '');
    ***REMOVED*** catch (e) {***REMOVED***
  ***REMOVED***
  return '';
***REMOVED***

/**
 * Creates an array of elements, sorted in ascending order by the results of
 * running each element in a collection thru each iteratee. This method
 * performs a stable sort, that is, it preserves the original sort order of
 * equal elements. The iteratees are invoked with one argument: (value).
 *
 * @static
 * @memberOf _
 * @since 0.1.0
 * @category Collection
 * @param {Array|Object***REMOVED*** collection The collection to iterate over.
 * @param {...(Function|Function[])***REMOVED*** [iteratees=[_.identity]]
 *  The iteratees to sort by.
 * @returns {Array***REMOVED*** Returns the new sorted array.
 * @example
 *
 * var users = [
 *   { 'user': 'fred',   'age': 48 ***REMOVED***,
 *   { 'user': 'barney', 'age': 36 ***REMOVED***,
 *   { 'user': 'fred',   'age': 40 ***REMOVED***,
 *   { 'user': 'barney', 'age': 34 ***REMOVED***
 * ];
 *
 * _.sortBy(users, function(o) { return o.user; ***REMOVED***);
 * // => objects for [['barney', 36], ['barney', 34], ['fred', 48], ['fred', 40]]
 *
 * _.sortBy(users, ['user', 'age']);
 * // => objects for [['barney', 34], ['barney', 36], ['fred', 40], ['fred', 48]]
 *
 * _.sortBy(users, 'user', function(o) {
 *   return Math.floor(o.age / 10);
 * ***REMOVED***);
 * // => objects for [['barney', 36], ['barney', 34], ['fred', 48], ['fred', 40]]
 */
var sortBy = baseRest(function(collection, iteratees) {
  if (collection == null) {
    return [];
  ***REMOVED***
  var length = iteratees.length;
  if (length > 1 && isIterateeCall(collection, iteratees[0], iteratees[1])) {
    iteratees = [];
  ***REMOVED*** else if (length > 2 && isIterateeCall(iteratees[0], iteratees[1], iteratees[2])) {
    iteratees = [iteratees[0]];
  ***REMOVED***
  return baseOrderBy(collection, baseFlatten(iteratees, 1), []);
***REMOVED***);

/**
 * Creates a function that memoizes the result of `func`. If `resolver` is
 * provided, it determines the cache key for storing the result based on the
 * arguments provided to the memoized function. By default, the first argument
 * provided to the memoized function is used as the map cache key. The `func`
 * is invoked with the `this` binding of the memoized function.
 *
 * **Note:** The cache is exposed as the `cache` property on the memoized
 * function. Its creation may be customized by replacing the `_.memoize.Cache`
 * constructor with one whose instances implement the
 * [`Map`](http://ecma-international.org/ecma-262/7.0/#sec-properties-of-the-map-prototype-object)
 * method interface of `delete`, `get`, `has`, and `set`.
 *
 * @static
 * @memberOf _
 * @since 0.1.0
 * @category Function
 * @param {Function***REMOVED*** func The function to have its output memoized.
 * @param {Function***REMOVED*** [resolver] The function to resolve the cache key.
 * @returns {Function***REMOVED*** Returns the new memoized function.
 * @example
 *
 * var object = { 'a': 1, 'b': 2 ***REMOVED***;
 * var other = { 'c': 3, 'd': 4 ***REMOVED***;
 *
 * var values = _.memoize(_.values);
 * values(object);
 * // => [1, 2]
 *
 * values(other);
 * // => [3, 4]
 *
 * object.a = 2;
 * values(object);
 * // => [1, 2]
 *
 * // Modify the result cache.
 * values.cache.set(object, ['a', 'b']);
 * values(object);
 * // => ['a', 'b']
 *
 * // Replace `_.memoize.Cache`.
 * _.memoize.Cache = WeakMap;
 */
function memoize(func, resolver) {
  if (typeof func != 'function' || (resolver && typeof resolver != 'function')) {
    throw new TypeError(FUNC_ERROR_TEXT);
  ***REMOVED***
  var memoized = function() {
    var args = arguments,
        key = resolver ? resolver.apply(this, args) : args[0],
        cache = memoized.cache;

    if (cache.has(key)) {
      return cache.get(key);
    ***REMOVED***
    var result = func.apply(this, args);
    memoized.cache = cache.set(key, result);
    return result;
  ***REMOVED***;
  memoized.cache = new (memoize.Cache || MapCache);
  return memoized;
***REMOVED***

// Assign cache to `_.memoize`.
memoize.Cache = MapCache;

/**
 * Performs a
 * [`SameValueZero`](http://ecma-international.org/ecma-262/7.0/#sec-samevaluezero)
 * comparison between two values to determine if they are equivalent.
 *
 * @static
 * @memberOf _
 * @since 4.0.0
 * @category Lang
 * @param {****REMOVED*** value The value to compare.
 * @param {****REMOVED*** other The other value to compare.
 * @returns {boolean***REMOVED*** Returns `true` if the values are equivalent, else `false`.
 * @example
 *
 * var object = { 'a': 1 ***REMOVED***;
 * var other = { 'a': 1 ***REMOVED***;
 *
 * _.eq(object, object);
 * // => true
 *
 * _.eq(object, other);
 * // => false
 *
 * _.eq('a', 'a');
 * // => true
 *
 * _.eq('a', Object('a'));
 * // => false
 *
 * _.eq(NaN, NaN);
 * // => true
 */
function eq(value, other) {
  return value === other || (value !== value && other !== other);
***REMOVED***

/**
 * Checks if `value` is likely an `arguments` object.
 *
 * @static
 * @memberOf _
 * @since 0.1.0
 * @category Lang
 * @param {****REMOVED*** value The value to check.
 * @returns {boolean***REMOVED*** Returns `true` if `value` is an `arguments` object,
 *  else `false`.
 * @example
 *
 * _.isArguments(function() { return arguments; ***REMOVED***());
 * // => true
 *
 * _.isArguments([1, 2, 3]);
 * // => false
 */
function isArguments(value) {
  // Safari 8.1 makes `arguments.callee` enumerable in strict mode.
  return isArrayLikeObject(value) && hasOwnProperty.call(value, 'callee') &&
    (!propertyIsEnumerable.call(value, 'callee') || objectToString.call(value) == argsTag);
***REMOVED***

/**
 * Checks if `value` is classified as an `Array` object.
 *
 * @static
 * @memberOf _
 * @since 0.1.0
 * @category Lang
 * @param {****REMOVED*** value The value to check.
 * @returns {boolean***REMOVED*** Returns `true` if `value` is an array, else `false`.
 * @example
 *
 * _.isArray([1, 2, 3]);
 * // => true
 *
 * _.isArray(document.body.children);
 * // => false
 *
 * _.isArray('abc');
 * // => false
 *
 * _.isArray(_.noop);
 * // => false
 */
var isArray = Array.isArray;

/**
 * Checks if `value` is array-like. A value is considered array-like if it's
 * not a function and has a `value.length` that's an integer greater than or
 * equal to `0` and less than or equal to `Number.MAX_SAFE_INTEGER`.
 *
 * @static
 * @memberOf _
 * @since 4.0.0
 * @category Lang
 * @param {****REMOVED*** value The value to check.
 * @returns {boolean***REMOVED*** Returns `true` if `value` is array-like, else `false`.
 * @example
 *
 * _.isArrayLike([1, 2, 3]);
 * // => true
 *
 * _.isArrayLike(document.body.children);
 * // => true
 *
 * _.isArrayLike('abc');
 * // => true
 *
 * _.isArrayLike(_.noop);
 * // => false
 */
function isArrayLike(value) {
  return value != null && isLength(value.length) && !isFunction(value);
***REMOVED***

/**
 * This method is like `_.isArrayLike` except that it also checks if `value`
 * is an object.
 *
 * @static
 * @memberOf _
 * @since 4.0.0
 * @category Lang
 * @param {****REMOVED*** value The value to check.
 * @returns {boolean***REMOVED*** Returns `true` if `value` is an array-like object,
 *  else `false`.
 * @example
 *
 * _.isArrayLikeObject([1, 2, 3]);
 * // => true
 *
 * _.isArrayLikeObject(document.body.children);
 * // => true
 *
 * _.isArrayLikeObject('abc');
 * // => false
 *
 * _.isArrayLikeObject(_.noop);
 * // => false
 */
function isArrayLikeObject(value) {
  return isObjectLike(value) && isArrayLike(value);
***REMOVED***

/**
 * Checks if `value` is classified as a `Function` object.
 *
 * @static
 * @memberOf _
 * @since 0.1.0
 * @category Lang
 * @param {****REMOVED*** value The value to check.
 * @returns {boolean***REMOVED*** Returns `true` if `value` is a function, else `false`.
 * @example
 *
 * _.isFunction(_);
 * // => true
 *
 * _.isFunction(/abc/);
 * // => false
 */
function isFunction(value) {
  // The use of `Object#toString` avoids issues with the `typeof` operator
  // in Safari 8-9 which returns 'object' for typed array and other constructors.
  var tag = isObject(value) ? objectToString.call(value) : '';
  return tag == funcTag || tag == genTag;
***REMOVED***

/**
 * Checks if `value` is a valid array-like length.
 *
 * **Note:** This method is loosely based on
 * [`ToLength`](http://ecma-international.org/ecma-262/7.0/#sec-tolength).
 *
 * @static
 * @memberOf _
 * @since 4.0.0
 * @category Lang
 * @param {****REMOVED*** value The value to check.
 * @returns {boolean***REMOVED*** Returns `true` if `value` is a valid length, else `false`.
 * @example
 *
 * _.isLength(3);
 * // => true
 *
 * _.isLength(Number.MIN_VALUE);
 * // => false
 *
 * _.isLength(Infinity);
 * // => false
 *
 * _.isLength('3');
 * // => false
 */
function isLength(value) {
  return typeof value == 'number' &&
    value > -1 && value % 1 == 0 && value <= MAX_SAFE_INTEGER;
***REMOVED***

/**
 * Checks if `value` is the
 * [language type](http://www.ecma-international.org/ecma-262/7.0/#sec-ecmascript-language-types)
 * of `Object`. (e.g. arrays, functions, objects, regexes, `new Number(0)`, and `new String('')`)
 *
 * @static
 * @memberOf _
 * @since 0.1.0
 * @category Lang
 * @param {****REMOVED*** value The value to check.
 * @returns {boolean***REMOVED*** Returns `true` if `value` is an object, else `false`.
 * @example
 *
 * _.isObject({***REMOVED***);
 * // => true
 *
 * _.isObject([1, 2, 3]);
 * // => true
 *
 * _.isObject(_.noop);
 * // => true
 *
 * _.isObject(null);
 * // => false
 */
function isObject(value) {
  var type = typeof value;
  return !!value && (type == 'object' || type == 'function');
***REMOVED***

/**
 * Checks if `value` is object-like. A value is object-like if it's not `null`
 * and has a `typeof` result of "object".
 *
 * @static
 * @memberOf _
 * @since 4.0.0
 * @category Lang
 * @param {****REMOVED*** value The value to check.
 * @returns {boolean***REMOVED*** Returns `true` if `value` is object-like, else `false`.
 * @example
 *
 * _.isObjectLike({***REMOVED***);
 * // => true
 *
 * _.isObjectLike([1, 2, 3]);
 * // => true
 *
 * _.isObjectLike(_.noop);
 * // => false
 *
 * _.isObjectLike(null);
 * // => false
 */
function isObjectLike(value) {
  return !!value && typeof value == 'object';
***REMOVED***

/**
 * Checks if `value` is classified as a `Symbol` primitive or object.
 *
 * @static
 * @memberOf _
 * @since 4.0.0
 * @category Lang
 * @param {****REMOVED*** value The value to check.
 * @returns {boolean***REMOVED*** Returns `true` if `value` is a symbol, else `false`.
 * @example
 *
 * _.isSymbol(Symbol.iterator);
 * // => true
 *
 * _.isSymbol('abc');
 * // => false
 */
function isSymbol(value) {
  return typeof value == 'symbol' ||
    (isObjectLike(value) && objectToString.call(value) == symbolTag);
***REMOVED***

/**
 * Checks if `value` is classified as a typed array.
 *
 * @static
 * @memberOf _
 * @since 3.0.0
 * @category Lang
 * @param {****REMOVED*** value The value to check.
 * @returns {boolean***REMOVED*** Returns `true` if `value` is a typed array, else `false`.
 * @example
 *
 * _.isTypedArray(new Uint8Array);
 * // => true
 *
 * _.isTypedArray([]);
 * // => false
 */
var isTypedArray = nodeIsTypedArray ? baseUnary(nodeIsTypedArray) : baseIsTypedArray;

/**
 * Converts `value` to a string. An empty string is returned for `null`
 * and `undefined` values. The sign of `-0` is preserved.
 *
 * @static
 * @memberOf _
 * @since 4.0.0
 * @category Lang
 * @param {****REMOVED*** value The value to process.
 * @returns {string***REMOVED*** Returns the string.
 * @example
 *
 * _.toString(null);
 * // => ''
 *
 * _.toString(-0);
 * // => '-0'
 *
 * _.toString([1, 2, 3]);
 * // => '1,2,3'
 */
function toString(value) {
  return value == null ? '' : baseToString(value);
***REMOVED***

/**
 * Gets the value at `path` of `object`. If the resolved value is
 * `undefined`, the `defaultValue` is returned in its place.
 *
 * @static
 * @memberOf _
 * @since 3.7.0
 * @category Object
 * @param {Object***REMOVED*** object The object to query.
 * @param {Array|string***REMOVED*** path The path of the property to get.
 * @param {****REMOVED*** [defaultValue] The value returned for `undefined` resolved values.
 * @returns {****REMOVED*** Returns the resolved value.
 * @example
 *
 * var object = { 'a': [{ 'b': { 'c': 3 ***REMOVED*** ***REMOVED***] ***REMOVED***;
 *
 * _.get(object, 'a[0].b.c');
 * // => 3
 *
 * _.get(object, ['a', '0', 'b', 'c']);
 * // => 3
 *
 * _.get(object, 'a.b.c', 'default');
 * // => 'default'
 */
function get(object, path, defaultValue) {
  var result = object == null ? undefined : baseGet(object, path);
  return result === undefined ? defaultValue : result;
***REMOVED***

/**
 * Checks if `path` is a direct or inherited property of `object`.
 *
 * @static
 * @memberOf _
 * @since 4.0.0
 * @category Object
 * @param {Object***REMOVED*** object The object to query.
 * @param {Array|string***REMOVED*** path The path to check.
 * @returns {boolean***REMOVED*** Returns `true` if `path` exists, else `false`.
 * @example
 *
 * var object = _.create({ 'a': _.create({ 'b': 2 ***REMOVED***) ***REMOVED***);
 *
 * _.hasIn(object, 'a');
 * // => true
 *
 * _.hasIn(object, 'a.b');
 * // => true
 *
 * _.hasIn(object, ['a', 'b']);
 * // => true
 *
 * _.hasIn(object, 'b');
 * // => false
 */
function hasIn(object, path) {
  return object != null && hasPath(object, path, baseHasIn);
***REMOVED***

/**
 * Creates an array of the own enumerable property names of `object`.
 *
 * **Note:** Non-object values are coerced to objects. See the
 * [ES spec](http://ecma-international.org/ecma-262/7.0/#sec-object.keys)
 * for more details.
 *
 * @static
 * @since 0.1.0
 * @memberOf _
 * @category Object
 * @param {Object***REMOVED*** object The object to query.
 * @returns {Array***REMOVED*** Returns the array of property names.
 * @example
 *
 * function Foo() {
 *   this.a = 1;
 *   this.b = 2;
 * ***REMOVED***
 *
 * Foo.prototype.c = 3;
 *
 * _.keys(new Foo);
 * // => ['a', 'b'] (iteration order is not guaranteed)
 *
 * _.keys('hi');
 * // => ['0', '1']
 */
function keys(object) {
  return isArrayLike(object) ? arrayLikeKeys(object) : baseKeys(object);
***REMOVED***

/**
 * This method returns the first argument it receives.
 *
 * @static
 * @since 0.1.0
 * @memberOf _
 * @category Util
 * @param {****REMOVED*** value Any value.
 * @returns {****REMOVED*** Returns `value`.
 * @example
 *
 * var object = { 'a': 1 ***REMOVED***;
 *
 * console.log(_.identity(object) === object);
 * // => true
 */
function identity(value) {
  return value;
***REMOVED***

/**
 * Creates a function that returns the value at `path` of a given object.
 *
 * @static
 * @memberOf _
 * @since 2.4.0
 * @category Util
 * @param {Array|string***REMOVED*** path The path of the property to get.
 * @returns {Function***REMOVED*** Returns the new accessor function.
 * @example
 *
 * var objects = [
 *   { 'a': { 'b': 2 ***REMOVED*** ***REMOVED***,
 *   { 'a': { 'b': 1 ***REMOVED*** ***REMOVED***
 * ];
 *
 * _.map(objects, _.property('a.b'));
 * // => [2, 1]
 *
 * _.map(_.sortBy(objects, _.property(['a', 'b'])), 'a.b');
 * // => [1, 2]
 */
function property(path) {
  return isKey(path) ? baseProperty(toKey(path)) : basePropertyDeep(path);
***REMOVED***

module.exports = sortBy;
