// workbox-strategies/NetworkOnly.js
/*
  Copyright 2018 Google LLC

  Use of this source code is governed by an MIT-style
  license that can be found in the LICENSE file or at
  https://opensource.org/licenses/MIT.
*/
import { assert ***REMOVED*** from 'workbox-core/_private/assert.js';
import { logger ***REMOVED*** from 'workbox-core/_private/logger.js';
import { timeout ***REMOVED*** from 'workbox-core/_private/timeout.js';
import { WorkboxError ***REMOVED*** from 'workbox-core/_private/WorkboxError.js';
import { Strategy ***REMOVED*** from './Strategy.js';
import { messages ***REMOVED*** from './utils/messages.js';
import './_version.js';
/**
 * An implementation of a
 * [network-only](https://developer.chrome.com/docs/workbox/caching-strategies-overview/#network-only)
 * request strategy.
 *
 * This class is useful if you want to take advantage of any
 * [Workbox plugins](https://developer.chrome.com/docs/workbox/using-plugins/).
 *
 * If the network request fails, this will throw a `WorkboxError` exception.
 *
 * @extends workbox-strategies.Strategy
 * @memberof workbox-strategies
 */
class NetworkOnly extends Strategy {
    /**
     * @param {Object***REMOVED*** [options]
     * @param {Array<Object>***REMOVED*** [options.plugins] [Plugins]{@link https://developers.google.com/web/tools/workbox/guides/using-plugins***REMOVED***
     * to use in conjunction with this caching strategy.
     * @param {Object***REMOVED*** [options.fetchOptions] Values passed along to the
     * [`init`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch#Parameters)
     * of [non-navigation](https://github.com/GoogleChrome/workbox/issues/1796)
     * `fetch()` requests made by this strategy.
     * @param {number***REMOVED*** [options.networkTimeoutSeconds] If set, any network requests
     * that fail to respond within the timeout will result in a network error.
     */
    constructor(options = {***REMOVED***) {
        super(options);
        this._networkTimeoutSeconds = options.networkTimeoutSeconds || 0;
    ***REMOVED***
    /**
     * @private
     * @param {Request|string***REMOVED*** request A request to run this strategy for.
     * @param {workbox-strategies.StrategyHandler***REMOVED*** handler The event that
     *     triggered the request.
     * @return {Promise<Response>***REMOVED***
     */
    async _handle(request, handler) {
        if (process.env.NODE_ENV !== 'production') {
            assert.isInstance(request, Request, {
                moduleName: 'workbox-strategies',
                className: this.constructor.name,
                funcName: '_handle',
                paramName: 'request',
            ***REMOVED***);
        ***REMOVED***
        let error = undefined;
        let response;
        try {
            const promises = [
                handler.fetch(request),
            ];
            if (this._networkTimeoutSeconds) {
                const timeoutPromise = timeout(this._networkTimeoutSeconds * 1000);
                promises.push(timeoutPromise);
            ***REMOVED***
            response = await Promise.race(promises);
            if (!response) {
                throw new Error(`Timed out the network response after ` +
                    `${this._networkTimeoutSeconds***REMOVED*** seconds.`);
            ***REMOVED***
        ***REMOVED***
        catch (err) {
            if (err instanceof Error) {
                error = err;
            ***REMOVED***
        ***REMOVED***
        if (process.env.NODE_ENV !== 'production') {
            logger.groupCollapsed(messages.strategyStart(this.constructor.name, request));
            if (response) {
                logger.log(`Got response from network.`);
            ***REMOVED***
            else {
                logger.log(`Unable to get a response from the network.`);
            ***REMOVED***
            messages.printFinalResponse(response);
            logger.groupEnd();
        ***REMOVED***
        if (!response) {
            throw new WorkboxError('no-response', { url: request.url, error ***REMOVED***);
        ***REMOVED***
        return response;
    ***REMOVED***
***REMOVED***
export { NetworkOnly ***REMOVED***;
