// yaml/browser/dist/nodes/Node.js
import { applyReviver ***REMOVED*** from '../doc/applyReviver.js';
import { NODE_TYPE, isDocument ***REMOVED*** from './identity.js';
import { toJS ***REMOVED*** from './toJS.js';

class NodeBase {
    constructor(type) {
        Object.defineProperty(this, NODE_TYPE, { value: type ***REMOVED***);
    ***REMOVED***
    /** Create a copy of this node.  */
    clone() {
        const copy = Object.create(Object.getPrototypeOf(this), Object.getOwnPropertyDescriptors(this));
        if (this.range)
            copy.range = this.range.slice();
        return copy;
    ***REMOVED***
    /** A plain JavaScript representation of this node. */
    toJS(doc, { mapAsMap, maxAliasCount, onAnchor, reviver ***REMOVED*** = {***REMOVED***) {
        if (!isDocument(doc))
            throw new TypeError('A document argument is required');
        const ctx = {
            anchors: new Map(),
            doc,
            keep: true,
            mapAsMap: mapAsMap === true,
            mapKeyWarned: false,
            maxAliasCount: typeof maxAliasCount === 'number' ? maxAliasCount : 100
        ***REMOVED***;
        const res = toJS(this, '', ctx);
        if (typeof onAnchor === 'function')
            for (const { count, res ***REMOVED*** of ctx.anchors.values())
                onAnchor(res, count);
        return typeof reviver === 'function'
            ? applyReviver(reviver, { '': res ***REMOVED***, '', res)
            : res;
    ***REMOVED***
***REMOVED***

export { NodeBase ***REMOVED***;
