// jsdom/lib/jsdom/living/helpers/create-event-accessor.js
"use strict";

const idlUtils = require("../generated/utils");
const ErrorEvent = require("../generated/ErrorEvent");
const EventHandlerNonNull = require("../generated/EventHandlerNonNull.js");
const OnBeforeUnloadEventHandlerNonNull = require("../generated/OnBeforeUnloadEventHandlerNonNull.js");
const OnErrorEventHandlerNonNull = require("../generated/OnErrorEventHandlerNonNull.js");
const reportException = require("./runtime-script-errors");

exports.appendHandler = (el, eventName) => {
  // tryImplForWrapper() is currently required due to use in Window.js
  idlUtils.tryImplForWrapper(el).addEventListener(eventName, event => {
    // https://html.spec.whatwg.org/#the-event-handler-processing-algorithm
    const callback = exports.getCurrentEventHandlerValue(el, eventName);
    if (callback === null) {
      return;
    ***REMOVED***

    const specialError = ErrorEvent.isImpl(event) && event.type === "error" &&
      event.currentTarget.constructor.name === "Window";

    let returnValue = null;
    // https://heycam.github.io/webidl/#es-invoking-callback-functions
    if (typeof callback === "function") {
      if (specialError) {
        returnValue = callback.call(
          event.currentTarget,
          event.message,
          event.filename,
          event.lineno,
          event.colno,
          event.error
        );
      ***REMOVED*** else {
        returnValue = callback.call(event.currentTarget, event);
      ***REMOVED***
    ***REMOVED***

    // TODO: we don't implement BeforeUnloadEvent so we can't brand-check here
    if (event.type === "beforeunload") {
      if (returnValue !== null) {
        event._canceledFlag = true;
        if (event.returnValue === "") {
          event.returnValue = returnValue;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED*** else if (specialError) {
      if (returnValue === true) {
        event._canceledFlag = true;
      ***REMOVED***
    ***REMOVED*** else if (returnValue === false) {
      event._canceledFlag = true;
    ***REMOVED***
  ***REMOVED***);
***REMOVED***;

// "Simple" in this case means "no content attributes involved"
exports.setupForSimpleEventAccessors = (prototype, events) => {
  prototype._getEventHandlerFor = function (event) {
    return this._eventHandlers ? this._eventHandlers[event] : undefined;
  ***REMOVED***;

  prototype._setEventHandlerFor = function (event, handler) {
    if (!this._registeredHandlers) {
      this._registeredHandlers = new Set();
      this._eventHandlers = Object.create(null);
    ***REMOVED***

    if (!this._registeredHandlers.has(event) && handler !== null) {
      this._registeredHandlers.add(event);
      exports.appendHandler(this, event);
    ***REMOVED***
    this._eventHandlers[event] = handler;
  ***REMOVED***;

  for (const event of events) {
    exports.createEventAccessor(prototype, event);
  ***REMOVED***
***REMOVED***;

// https://html.spec.whatwg.org/multipage/webappapis.html#getting-the-current-value-of-the-event-handler
exports.getCurrentEventHandlerValue = (target, event) => {
  const value = target._getEventHandlerFor(event);
  if (!value) {
    return null;
  ***REMOVED***

  if (value.body !== undefined) {
    let element, document, fn;
    if (target.constructor.name === "Window") {
      element = null;
      document = idlUtils.implForWrapper(target.document);
    ***REMOVED*** else {
      element = target;
      document = element.ownerDocument;
    ***REMOVED***
    const { body ***REMOVED*** = value;

    const formOwner = element !== null && element.form ? element.form : null;
    const window = target.constructor.name === "Window" && target._document ? target : document.defaultView;

    try {
      // eslint-disable-next-line no-new-func
      Function(body); // properly error out on syntax errors
      // Note: this won't execute body; that would require `Function(body)()`.
    ***REMOVED*** catch (e) {
      if (window) {
        reportException(window, e);
      ***REMOVED***
      target._setEventHandlerFor(event, null);
      return null;
    ***REMOVED***

    // Note: the with (window) { ***REMOVED*** is not necessary in Node, but is necessary in a browserified environment.

    const createFunction = document.defaultView.Function;
    if (event === "error" && element === null) {
      const sourceURL = document ? `\n//# sourceURL=${document.URL***REMOVED***` : "";

      fn = createFunction(`\
with (arguments[0]) { return function onerror(event, source, lineno, colno, error) {
${body***REMOVED***
***REMOVED***; ***REMOVED***${sourceURL***REMOVED***`)(window);

      fn = OnErrorEventHandlerNonNull.convert(fn);
    ***REMOVED*** else {
      const calls = [];
      if (element !== null) {
        calls.push(idlUtils.wrapperForImpl(document));
      ***REMOVED***

      if (formOwner !== null) {
        calls.push(idlUtils.wrapperForImpl(formOwner));
      ***REMOVED***

      if (element !== null) {
        calls.push(idlUtils.wrapperForImpl(element));
      ***REMOVED***

      let wrapperBody = `\
with (arguments[0]) { return function on${event***REMOVED***(event) {
${body***REMOVED***
***REMOVED***; ***REMOVED***`;

      // eslint-disable-next-line no-unused-vars
      for (const call of calls) {
        wrapperBody = `\
with (arguments[0]) { return function () {
${wrapperBody***REMOVED***
***REMOVED***; ***REMOVED***`;
      ***REMOVED***

      if (document) {
        wrapperBody += `\n//# sourceURL=${document.URL***REMOVED***`;
      ***REMOVED***

      fn = createFunction(wrapperBody)(window);
      for (const call of calls) {
        fn = fn(call);
      ***REMOVED***

      if (event === "beforeunload") {
        fn = OnBeforeUnloadEventHandlerNonNull.convert(fn);
      ***REMOVED*** else {
        fn = EventHandlerNonNull.convert(fn);
      ***REMOVED***
    ***REMOVED***

    target._setEventHandlerFor(event, fn);
  ***REMOVED***

  return target._getEventHandlerFor(event);
***REMOVED***;

// https://html.spec.whatwg.org/multipage/webappapis.html#event-handler-idl-attributes
// TODO: Consider replacing this with `[ReflectEvent]`
exports.createEventAccessor = (obj, event) => {
  Object.defineProperty(obj, "on" + event, {
    configurable: true,
    enumerable: true,
    get() {
      return exports.getCurrentEventHandlerValue(this, event);
    ***REMOVED***,
    set(val) {
      this._setEventHandlerFor(event, val);
    ***REMOVED***
  ***REMOVED***);
***REMOVED***;
