// jsdom/lib/jsdom/living/helpers/runtime-script-errors.js
"use strict";
const util = require("util");
const idlUtils = require("../generated/utils");
const ErrorEvent = require("../generated/ErrorEvent");
const { createAnEvent ***REMOVED*** = require("../helpers/events");

const errorReportingMode = Symbol("error reporting mode");

// https://html.spec.whatwg.org/multipage/webappapis.html#report-the-error
// Omits script parameter and any check for muted errors.
// Takes target as an EventTarget impl.
// Takes error object, message, and location as params, unlike the spec.
// Returns whether the event was handled or not.
function reportAnError(line, col, target, errorObject, message, location) {
  if (target[errorReportingMode]) {
    return false;
  ***REMOVED***

  target[errorReportingMode] = true;

  if (typeof message !== "string") {
    message = "uncaught exception: " + util.inspect(errorObject);
  ***REMOVED***

  const event = createAnEvent("error", target._globalObject, ErrorEvent, {
    cancelable: true,
    message,
    filename: location,
    lineno: line,
    colno: col,
    error: errorObject
  ***REMOVED***);

  try {
    target._dispatch(event);
  ***REMOVED*** finally {
    target[errorReportingMode] = false;
    return event.defaultPrevented;
  ***REMOVED***
***REMOVED***

module.exports = function reportException(window, error, filenameHint) {
  // This function will give good results on real Error objects with stacks; poor ones otherwise

  const stack = error && error.stack;
  const lines = stack && stack.split("\n");

  // Find the first line that matches; important for multi-line messages
  let pieces;
  if (lines) {
    for (let i = 1; i < lines.length && !pieces; ++i) {
      pieces = lines[i].match(/at (?:(.+)\s+)?\(?(?:(.+?):(\d+):(\d+)|([^)]+))\)?/);
    ***REMOVED***
  ***REMOVED***

  const fileName = (pieces && pieces[2]) || filenameHint || window._document.URL;
  const lineNumber = (pieces && parseInt(pieces[3])) || 0;
  const columnNumber = (pieces && parseInt(pieces[4])) || 0;

  const windowImpl = idlUtils.implForWrapper(window);

  const handled = reportAnError(lineNumber, columnNumber, windowImpl, error, error && error.message, fileName);

  if (!handled) {
    const errorString = shouldBeDisplayedAsError(error) ? `[${error.name***REMOVED***: ${error.message***REMOVED***]` : util.inspect(error);
    const jsdomError = new Error(`Uncaught ${errorString***REMOVED***`);
    jsdomError.detail = error;
    jsdomError.type = "unhandled exception";

    window._virtualConsole.emit("jsdomError", jsdomError);
  ***REMOVED***
***REMOVED***;

function shouldBeDisplayedAsError(x) {
  return x && x.name && x.message !== undefined && x.stack;
***REMOVED***
