// css-declaration-sorter/src/main.mjs
import { shorthandData ***REMOVED*** from './shorthand-data.mjs';
import { bubbleSort ***REMOVED*** from './bubble-sort.mjs';

const builtInOrders = [
  'alphabetical',
  'concentric-css',
  'smacss',
];

export const cssDeclarationSorter = ({ order = 'alphabetical', keepOverrides = false ***REMOVED*** = {***REMOVED***) => ({
  postcssPlugin: 'css-declaration-sorter',
  OnceExit (css) {
    let withKeepOverrides = comparator => comparator;
    if (keepOverrides) {
      withKeepOverrides = withOverridesComparator(shorthandData);
    ***REMOVED***

    if (typeof order === 'function') {
      return processCss({ css, comparator: withKeepOverrides(order) ***REMOVED***);
    ***REMOVED***

    if (!builtInOrders.includes(order))
      return Promise.reject(
        Error([
          `Invalid built-in order '${order***REMOVED***' provided.`,
          `Available built-in orders are: ${builtInOrders***REMOVED***`,
        ].join('\n'))
      );

    return import(`../orders/${order***REMOVED***.mjs`)
      .then(({ properties ***REMOVED***) => processCss({
        css,
        comparator: withKeepOverrides(orderComparator(properties)),
      ***REMOVED***));
  ***REMOVED***,
***REMOVED***);

cssDeclarationSorter.postcss = true;

// Kept for backward compatibility
export default cssDeclarationSorter;

function processCss ({ css, comparator ***REMOVED***) {
  const comments = [];
  const rulesCache = [];

  css.walk(node => {
    const nodes = node.nodes;
    const type = node.type;

    if (type === 'comment') {
      // Don't do anything to root comments or the last newline comment
      const isNewlineNode = node.raws.before && node.raws.before.includes('\n');
      const lastNewlineNode = isNewlineNode && !node.next();
      const onlyNode = !node.prev() && !node.next() || !node.parent;

      if (lastNewlineNode || onlyNode || node.parent.type === 'root') {
        return;
      ***REMOVED***

      if (isNewlineNode) {
        const pairedNode = node.next() || node.prev();
        if (pairedNode) {
          comments.unshift({
            'comment': node,
            'pairedNode': pairedNode,
            'insertPosition': node.next() ? 'Before' : 'After',
          ***REMOVED***);
          node.remove();
        ***REMOVED***
      ***REMOVED*** else {
        const pairedNode = node.prev() || node.next();
        if (pairedNode) {
          comments.push({
            'comment': node,
            'pairedNode': pairedNode,
            'insertPosition': 'After',
          ***REMOVED***);
          node.remove();
        ***REMOVED***
      ***REMOVED***
      return;
    ***REMOVED***

    // Add rule-like nodes to a cache so that we can remove all
    // comment nodes before we start sorting.
    const isRule = type === 'rule' || type === 'atrule';
    if (isRule && nodes && nodes.length > 1) {
      rulesCache.push(nodes);
    ***REMOVED***
  ***REMOVED***);

  // Perform a sort once all comment nodes are removed
  rulesCache.forEach(nodes => {
    sortCssDeclarations({ nodes, comparator ***REMOVED***);
  ***REMOVED***);

  // Add comments back to the nodes they are paired with
  comments.forEach(node => {
    const pairedNode = node.pairedNode;
    node.comment.remove();
    pairedNode.parent && pairedNode.parent['insert' + node.insertPosition](pairedNode, node.comment);
  ***REMOVED***);
***REMOVED***

function sortCssDeclarations ({ nodes, comparator ***REMOVED***) {
  bubbleSort(nodes, (a, b) => {
    if (a.type === 'decl' && b.type === 'decl') {
      return comparator(a.prop, b.prop);
    ***REMOVED*** else {
      return compareDifferentType(a, b);
    ***REMOVED***
  ***REMOVED***);
***REMOVED***

function withOverridesComparator (shorthandData) {
  return function (comparator) {
    return function (a, b) {
      a = removeVendorPrefix(a);
      b = removeVendorPrefix(b);

      if (shorthandData[a] && shorthandData[a].includes(b)) return 0;
      if (shorthandData[b] && shorthandData[b].includes(a)) return 0;

      return comparator(a, b);
    ***REMOVED***;
  ***REMOVED***;
***REMOVED***

function orderComparator (order) {
  return function (a, b) {
    return order.indexOf(a) - order.indexOf(b);
  ***REMOVED***;
***REMOVED***

function compareDifferentType (a, b) {
  if (b.type === 'atrule' || a.type === 'atrule') {
    return 0;
  ***REMOVED***

  return a.type === 'decl' ? -1 : b.type === 'decl' ? 1 : 0;
***REMOVED***

function removeVendorPrefix (property) {
  return property.replace(/^-\w+-/, '');
***REMOVED***
