// yaml/browser/dist/schema/yaml-1.1/omap.js
import { isScalar, isPair ***REMOVED*** from '../../nodes/identity.js';
import { toJS ***REMOVED*** from '../../nodes/toJS.js';
import { YAMLMap ***REMOVED*** from '../../nodes/YAMLMap.js';
import { YAMLSeq ***REMOVED*** from '../../nodes/YAMLSeq.js';
import { resolvePairs, createPairs ***REMOVED*** from './pairs.js';

class YAMLOMap extends YAMLSeq {
    constructor() {
        super();
        this.add = YAMLMap.prototype.add.bind(this);
        this.delete = YAMLMap.prototype.delete.bind(this);
        this.get = YAMLMap.prototype.get.bind(this);
        this.has = YAMLMap.prototype.has.bind(this);
        this.set = YAMLMap.prototype.set.bind(this);
        this.tag = YAMLOMap.tag;
    ***REMOVED***
    /**
     * If `ctx` is given, the return type is actually `Map<unknown, unknown>`,
     * but TypeScript won't allow widening the signature of a child method.
     */
    toJSON(_, ctx) {
        if (!ctx)
            return super.toJSON(_);
        const map = new Map();
        if (ctx?.onCreate)
            ctx.onCreate(map);
        for (const pair of this.items) {
            let key, value;
            if (isPair(pair)) {
                key = toJS(pair.key, '', ctx);
                value = toJS(pair.value, key, ctx);
            ***REMOVED***
            else {
                key = toJS(pair, '', ctx);
            ***REMOVED***
            if (map.has(key))
                throw new Error('Ordered maps must not include duplicate keys');
            map.set(key, value);
        ***REMOVED***
        return map;
    ***REMOVED***
    static from(schema, iterable, ctx) {
        const pairs = createPairs(schema, iterable, ctx);
        const omap = new this();
        omap.items = pairs.items;
        return omap;
    ***REMOVED***
***REMOVED***
YAMLOMap.tag = 'tag:yaml.org,2002:omap';
const omap = {
    collection: 'seq',
    identify: value => value instanceof Map,
    nodeClass: YAMLOMap,
    default: false,
    tag: 'tag:yaml.org,2002:omap',
    resolve(seq, onError) {
        const pairs = resolvePairs(seq, onError);
        const seenKeys = [];
        for (const { key ***REMOVED*** of pairs.items) {
            if (isScalar(key)) {
                if (seenKeys.includes(key.value)) {
                    onError(`Ordered maps must not include duplicate keys: ${key.value***REMOVED***`);
                ***REMOVED***
                else {
                    seenKeys.push(key.value);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        return Object.assign(new YAMLOMap(), pairs);
    ***REMOVED***,
    createNode: (schema, iterable, ctx) => YAMLOMap.from(schema, iterable, ctx)
***REMOVED***;

export { YAMLOMap, omap ***REMOVED***;
