// css-select-base-adapter/index.js
'use strict'

module.exports = adapterFactory;

function adapterFactory(implementation){
	ensureImplementation(implementation);

	var adapter = {***REMOVED***

	var baseAdapter = {
		removeSubsets: function (nodes){
			return removeSubsets(adapter, nodes);
		***REMOVED***,
		existsOne: function(test, elems){
			return existsOne(adapter, test, elems);
		***REMOVED***,
		getSiblings: function(elem){
			return getSiblings(adapter, elem);
		***REMOVED***,
		hasAttrib: function(elem, name){
			return hasAttrib(adapter, elem, name);
		***REMOVED***,
		findOne: function(test, arr){
			return findOne(adapter, test, arr);
		***REMOVED***,
		findAll: function(test, elems){
			return findAll(adapter, test, elems)
		***REMOVED***
	***REMOVED***;

	Object.assign(adapter, baseAdapter, implementation);

	return adapter;
***REMOVED***

var expectImplemented = [
	"isTag", "getAttributeValue", "getChildren", "getName", "getParent",
	"getText"
];

function ensureImplementation(implementation){
	if(!implementation)	throw new TypeError("Expected implementation")

	var notImplemented = expectImplemented.filter(function(fname){
		return typeof implementation[fname] !== "function";
	***REMOVED***);

	if(notImplemented.length){
		var notList = "(" + notImplemented.join(", ") + ")";
		var message = "Expected functions " + notList + " to be implemented";
		throw new Error(message);
	***REMOVED***
***REMOVED***

function removeSubsets(adapter, nodes){
	var idx = nodes.length, node, ancestor, replace;

	// Check if each node (or one of its ancestors) is already contained in the
	// array.
	while(--idx > -1){
		node = ancestor = nodes[idx];

		// Temporarily remove the node under consideration
		nodes[idx] = null;
		replace = true;

		while(ancestor){
			if(nodes.indexOf(ancestor) > -1){
				replace = false;
				nodes.splice(idx, 1);
				break;
			***REMOVED***
			ancestor = adapter.getParent(ancestor)
		***REMOVED***

		// If the node has been found to be unique, re-insert it.
		if(replace){
			nodes[idx] = node;
		***REMOVED***
	***REMOVED***

	return nodes;
***REMOVED***

function existsOne(adapter, test, elems){
	return elems.some(function(elem){
		return adapter.isTag(elem) ?
			test(elem) || adapter.existsOne(test, adapter.getChildren(elem)) :
			false;
	***REMOVED***);
***REMOVED***

function getSiblings(adapter, elem){
	var parent = adapter.getParent(elem);
	return parent && adapter.getChildren(parent);
***REMOVED***


function hasAttrib(adapter, elem, name){
	return adapter.getAttributeValue(elem,name) !== undefined
***REMOVED***

function findOne(adapter, test, arr){
	var elem = null;

	for(var i = 0, l = arr.length; i < l && !elem; i++){
		if(test(arr[i])){
			elem = arr[i];
		***REMOVED*** else {
			var childs = adapter.getChildren(arr[i]);
			if(childs && childs.length > 0){
				elem = adapter.findOne(test, childs);
			***REMOVED***
		***REMOVED***
	***REMOVED***

	return elem;
***REMOVED***

function findAll(adapter, test, elems){
	var result = [];

	for(var i = 0, j = elems.length; i < j; i++){
		if(!adapter.isTag(elems[i])) continue;
		if(test(elems[i])) result.push(elems[i]);
		var childs = adapter.getChildren(elems[i]);
		if(childs) result = result.concat(adapter.findAll(test, childs));
	***REMOVED***

	return result;
***REMOVED***
