// css-tree/lib/syntax/node/Nth.js
module.exports = {
    name: 'Nth',
    structure: {
        nth: ['AnPlusB', 'Identifier'],
        selector: ['SelectorList', null]
    ***REMOVED***,
    parse: function(allowOfClause) {
        this.scanner.skipSC();

        var start = this.scanner.tokenStart;
        var end = start;
        var selector = null;
        var query;

        if (this.scanner.lookupValue(0, 'odd') || this.scanner.lookupValue(0, 'even')) {
            query = this.Identifier();
        ***REMOVED*** else {
            query = this.AnPlusB();
        ***REMOVED***

        this.scanner.skipSC();

        if (allowOfClause && this.scanner.lookupValue(0, 'of')) {
            this.scanner.next();

            selector = this.SelectorList();

            if (this.needPositions) {
                end = this.getLastListNode(selector.children).loc.end.offset;
            ***REMOVED***
        ***REMOVED*** else {
            if (this.needPositions) {
                end = query.loc.end.offset;
            ***REMOVED***
        ***REMOVED***

        return {
            type: 'Nth',
            loc: this.getLocation(start, end),
            nth: query,
            selector: selector
        ***REMOVED***;
    ***REMOVED***,
    generate: function(node) {
        this.node(node.nth);
        if (node.selector !== null) {
            this.chunk(' of ');
            this.node(node.selector);
        ***REMOVED***
    ***REMOVED***
***REMOVED***;
