// @babel/traverse/lib/path/lib/virtual-types-validator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.isBindingIdentifier = isBindingIdentifier;
exports.isBlockScoped = isBlockScoped;
exports.isExpression = isExpression;
exports.isFlow = isFlow;
exports.isForAwaitStatement = isForAwaitStatement;
exports.isGenerated = isGenerated;
exports.isPure = isPure;
exports.isReferenced = isReferenced;
exports.isReferencedIdentifier = isReferencedIdentifier;
exports.isReferencedMemberExpression = isReferencedMemberExpression;
exports.isRestProperty = isRestProperty;
exports.isScope = isScope;
exports.isSpreadProperty = isSpreadProperty;
exports.isStatement = isStatement;
exports.isUser = isUser;
exports.isVar = isVar;
var _t = require("@babel/types");
const {
  isBinding,
  isBlockScoped: nodeIsBlockScoped,
  isExportDeclaration,
  isExpression: nodeIsExpression,
  isFlow: nodeIsFlow,
  isForStatement,
  isForXStatement,
  isIdentifier,
  isImportDeclaration,
  isImportSpecifier,
  isJSXIdentifier,
  isJSXMemberExpression,
  isMemberExpression,
  isRestElement: nodeIsRestElement,
  isReferenced: nodeIsReferenced,
  isScope: nodeIsScope,
  isStatement: nodeIsStatement,
  isVar: nodeIsVar,
  isVariableDeclaration,
  react,
  isForOfStatement
***REMOVED*** = _t;
const {
  isCompatTag
***REMOVED*** = react;
function isReferencedIdentifier(opts) {
  const {
    node,
    parent
  ***REMOVED*** = this;
  if (!isIdentifier(node, opts) && !isJSXMemberExpression(parent, opts)) {
    if (isJSXIdentifier(node, opts)) {
      if (isCompatTag(node.name)) return false;
    ***REMOVED*** else {
      return false;
    ***REMOVED***
  ***REMOVED***
  return nodeIsReferenced(node, parent, this.parentPath.parent);
***REMOVED***
function isReferencedMemberExpression() {
  const {
    node,
    parent
  ***REMOVED*** = this;
  return isMemberExpression(node) && nodeIsReferenced(node, parent);
***REMOVED***
function isBindingIdentifier() {
  const {
    node,
    parent
  ***REMOVED*** = this;
  const grandparent = this.parentPath.parent;
  return isIdentifier(node) && isBinding(node, parent, grandparent);
***REMOVED***
function isStatement() {
  const {
    node,
    parent
  ***REMOVED*** = this;
  if (nodeIsStatement(node)) {
    if (isVariableDeclaration(node)) {
      if (isForXStatement(parent, {
        left: node
      ***REMOVED***)) return false;
      if (isForStatement(parent, {
        init: node
      ***REMOVED***)) return false;
    ***REMOVED***
    return true;
  ***REMOVED*** else {
    return false;
  ***REMOVED***
***REMOVED***
function isExpression() {
  if (this.isIdentifier()) {
    return this.isReferencedIdentifier();
  ***REMOVED*** else {
    return nodeIsExpression(this.node);
  ***REMOVED***
***REMOVED***
function isScope() {
  return nodeIsScope(this.node, this.parent);
***REMOVED***
function isReferenced() {
  return nodeIsReferenced(this.node, this.parent);
***REMOVED***
function isBlockScoped() {
  return nodeIsBlockScoped(this.node);
***REMOVED***
function isVar() {
  return nodeIsVar(this.node);
***REMOVED***
function isUser() {
  return this.node && !!this.node.loc;
***REMOVED***
function isGenerated() {
  return !this.isUser();
***REMOVED***
function isPure(constantsOnly) {
  return this.scope.isPure(this.node, constantsOnly);
***REMOVED***
function isFlow() {
  const {
    node
  ***REMOVED*** = this;
  if (nodeIsFlow(node)) {
    return true;
  ***REMOVED*** else if (isImportDeclaration(node)) {
    return node.importKind === "type" || node.importKind === "typeof";
  ***REMOVED*** else if (isExportDeclaration(node)) {
    return node.exportKind === "type";
  ***REMOVED*** else if (isImportSpecifier(node)) {
    return node.importKind === "type" || node.importKind === "typeof";
  ***REMOVED*** else {
    return false;
  ***REMOVED***
***REMOVED***
function isRestProperty() {
  var _this$parentPath;
  return nodeIsRestElement(this.node) && ((_this$parentPath = this.parentPath) == null ? void 0 : _this$parentPath.isObjectPattern());
***REMOVED***
function isSpreadProperty() {
  var _this$parentPath2;
  return nodeIsRestElement(this.node) && ((_this$parentPath2 = this.parentPath) == null ? void 0 : _this$parentPath2.isObjectExpression());
***REMOVED***
function isForAwaitStatement() {
  return isForOfStatement(this.node, {
    await: true
  ***REMOVED***);
***REMOVED***
{
  exports.isExistentialTypeParam = function isExistentialTypeParam() {
    throw new Error("`path.isExistentialTypeParam` has been renamed to `path.isExistsTypeAnnotation()` in Babel 7.");
  ***REMOVED***;
  exports.isNumericLiteralTypeAnnotation = function isNumericLiteralTypeAnnotation() {
    throw new Error("`path.isNumericLiteralTypeAnnotation()` has been renamed to `path.isNumberLiteralTypeAnnotation()` in Babel 7.");
  ***REMOVED***;
***REMOVED***
