// eslint-scope/lib/reference.js
/*
  Copyright (C) 2015 Yusuke Suzuki <utatane.tea@gmail.com>

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
  ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

const READ = 0x1;
const WRITE = 0x2;
const RW = READ | WRITE;

/**
 * A Reference represents a single occurrence of an identifier in code.
 * @constructor Reference
 */
class Reference {
    constructor(ident, scope, flag, writeExpr, maybeImplicitGlobal, partial, init) {

        /**
         * Identifier syntax node.
         * @member {espreeIdentifier***REMOVED*** Reference#identifier
         */
        this.identifier = ident;

        /**
         * Reference to the enclosing Scope.
         * @member {Scope***REMOVED*** Reference#from
         */
        this.from = scope;

        /**
         * Whether the reference comes from a dynamic scope (such as 'eval',
         * 'with', etc.), and may be trapped by dynamic scopes.
         * @member {boolean***REMOVED*** Reference#tainted
         */
        this.tainted = false;

        /**
         * The variable this reference is resolved with.
         * @member {Variable***REMOVED*** Reference#resolved
         */
        this.resolved = null;

        /**
         * The read-write mode of the reference. (Value is one of {@link
         * Reference.READ***REMOVED***, {@link Reference.RW***REMOVED***, {@link Reference.WRITE***REMOVED***).
         * @member {number***REMOVED*** Reference#flag
         * @private
         */
        this.flag = flag;
        if (this.isWrite()) {

            /**
             * If reference is writeable, this is the tree being written to it.
             * @member {espreeNode***REMOVED*** Reference#writeExpr
             */
            this.writeExpr = writeExpr;

            /**
             * Whether the Reference might refer to a partial value of writeExpr.
             * @member {boolean***REMOVED*** Reference#partial
             */
            this.partial = partial;

            /**
             * Whether the Reference is to write of initialization.
             * @member {boolean***REMOVED*** Reference#init
             */
            this.init = init;
        ***REMOVED***
        this.__maybeImplicitGlobal = maybeImplicitGlobal;
    ***REMOVED***

    /**
     * Whether the reference is static.
     * @function Reference#isStatic
     * @returns {boolean***REMOVED*** static
     */
    isStatic() {
        return !this.tainted && this.resolved && this.resolved.scope.isStatic();
    ***REMOVED***

    /**
     * Whether the reference is writeable.
     * @function Reference#isWrite
     * @returns {boolean***REMOVED*** write
     */
    isWrite() {
        return !!(this.flag & Reference.WRITE);
    ***REMOVED***

    /**
     * Whether the reference is readable.
     * @function Reference#isRead
     * @returns {boolean***REMOVED*** read
     */
    isRead() {
        return !!(this.flag & Reference.READ);
    ***REMOVED***

    /**
     * Whether the reference is read-only.
     * @function Reference#isReadOnly
     * @returns {boolean***REMOVED*** read only
     */
    isReadOnly() {
        return this.flag === Reference.READ;
    ***REMOVED***

    /**
     * Whether the reference is write-only.
     * @function Reference#isWriteOnly
     * @returns {boolean***REMOVED*** write only
     */
    isWriteOnly() {
        return this.flag === Reference.WRITE;
    ***REMOVED***

    /**
     * Whether the reference is read-write.
     * @function Reference#isReadWrite
     * @returns {boolean***REMOVED*** read write
     */
    isReadWrite() {
        return this.flag === Reference.RW;
    ***REMOVED***
***REMOVED***

/**
 * @constant Reference.READ
 * @private
 */
Reference.READ = READ;

/**
 * @constant Reference.WRITE
 * @private
 */
Reference.WRITE = WRITE;

/**
 * @constant Reference.RW
 * @private
 */
Reference.RW = RW;

export default Reference;

/* vim: set sw=4 ts=4 et tw=80 : */
