// babel-plugin-transform-react-remove-prop-types/lib/isStatelessComponent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = isStatelessComponent;
var traversed = Symbol('traversed');

function isJSXElementOrReactCreateElement(path) {
  var visited = false;
  path.traverse({
    CallExpression: function CallExpression(path2) {
      var callee = path2.get('callee');

      if (callee.matchesPattern('React.createElement') || callee.matchesPattern('React.cloneElement') || callee.node.name === 'cloneElement') {
        visited = true;
      ***REMOVED***
    ***REMOVED***,
    JSXElement: function JSXElement() {
      visited = true;
    ***REMOVED***
  ***REMOVED***);
  return visited;
***REMOVED***

function isReturningJSXElement(path) {
  var iteration = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;

  // Early exit for ArrowFunctionExpressions, there is no ReturnStatement node.
  if (path.node.init && path.node.init.body && isJSXElementOrReactCreateElement(path)) {
    return true;
  ***REMOVED***

  if (iteration > 20) {
    throw new Error('transform-react-remove-prop-type: infinite loop detected.');
  ***REMOVED***

  var visited = false;
  path.traverse({
    ReturnStatement: function ReturnStatement(path2) {
      // We have already found what we are looking for.
      if (visited) {
        return;
      ***REMOVED***

      var argument = path2.get('argument'); // Nothing is returned

      if (!argument.node) {
        return;
      ***REMOVED***

      if (isJSXElementOrReactCreateElement(path2)) {
        visited = true;
        return;
      ***REMOVED***

      if (argument.node.type === 'CallExpression') {
        var name = argument.get('callee').node.name;
        var binding = path.scope.getBinding(name);

        if (!binding) {
          return;
        ***REMOVED*** // Prevents infinite traverse loop.


        if (binding.path[traversed]) {
          return;
        ***REMOVED***

        binding.path[traversed] = true;

        if (isReturningJSXElement(binding.path, iteration + 1)) {
          visited = true;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***);
  return visited;
***REMOVED***

var VALID_POSSIBLE_STATELESS_COMPONENT_TYPES = ['VariableDeclarator', 'FunctionDeclaration']; // Returns `true` if the path represents a function which returns a JSXElement

function isStatelessComponent(path) {
  if (VALID_POSSIBLE_STATELESS_COMPONENT_TYPES.indexOf(path.node.type) === -1) {
    return false;
  ***REMOVED***

  if (isReturningJSXElement(path)) {
    return true;
  ***REMOVED***

  return false;
***REMOVED***
