// jest-haste-map/build/watchers/common.js
// vendored from https://github.com/amasad/sane/blob/64ff3a870c42e84f744086884bf55a4f9c22d376/src/common.js
'use strict';

const platform = require('os').platform();

const path = require('path');

const anymatch = require('anymatch');

const micromatch = require('micromatch');

const walker = require('walker');
/**
 * Constants
 */

exports.DEFAULT_DELAY = 100;
exports.CHANGE_EVENT = 'change';
exports.DELETE_EVENT = 'delete';
exports.ADD_EVENT = 'add';
exports.ALL_EVENT = 'all';
/**
 * Assigns options to the watcher.
 *
 * @param {NodeWatcher|PollWatcher|WatchmanWatcher***REMOVED*** watcher
 * @param {?object***REMOVED*** opts
 * @return {boolean***REMOVED***
 * @public
 */

exports.assignOptions = function (watcher, opts) {
  opts = opts || {***REMOVED***;
  watcher.globs = opts.glob || [];
  watcher.dot = opts.dot || false;
  watcher.ignored = opts.ignored || false;

  if (!Array.isArray(watcher.globs)) {
    watcher.globs = [watcher.globs];
  ***REMOVED***

  watcher.hasIgnore =
    Boolean(opts.ignored) && !(Array.isArray(opts) && opts.length > 0);
  watcher.doIgnore = opts.ignored ? anymatch(opts.ignored) : () => false;

  if (opts.watchman && opts.watchmanPath) {
    watcher.watchmanPath = opts.watchmanPath;
  ***REMOVED***

  return opts;
***REMOVED***;
/**
 * Checks a file relative path against the globs array.
 *
 * @param {array***REMOVED*** globs
 * @param {string***REMOVED*** relativePath
 * @return {boolean***REMOVED***
 * @public
 */

exports.isFileIncluded = function (globs, dot, doIgnore, relativePath) {
  if (doIgnore(relativePath)) {
    return false;
  ***REMOVED***

  return globs.length
    ? micromatch.some(relativePath, globs, {
        dot
      ***REMOVED***)
    : dot || micromatch.some(relativePath, '**/*');
***REMOVED***;
/**
 * Traverse a directory recursively calling `callback` on every directory.
 *
 * @param {string***REMOVED*** dir
 * @param {function***REMOVED*** dirCallback
 * @param {function***REMOVED*** fileCallback
 * @param {function***REMOVED*** endCallback
 * @param {****REMOVED*** ignored
 * @public
 */

exports.recReaddir = function (
  dir,
  dirCallback,
  fileCallback,
  endCallback,
  errorCallback,
  ignored
) {
  walker(dir)
    .filterDir(currentDir => !anymatch(ignored, currentDir))
    .on('dir', normalizeProxy(dirCallback))
    .on('file', normalizeProxy(fileCallback))
    .on('error', errorCallback)
    .on('end', () => {
      if (platform === 'win32') {
        setTimeout(endCallback, 1000);
      ***REMOVED*** else {
        endCallback();
      ***REMOVED***
    ***REMOVED***);
***REMOVED***;
/**
 * Returns a callback that when called will normalize a path and call the
 * original callback
 *
 * @param {function***REMOVED*** callback
 * @return {function***REMOVED***
 * @private
 */

function normalizeProxy(callback) {
  return (filepath, stats) => callback(path.normalize(filepath), stats);
***REMOVED***
