// @babel/traverse/lib/context.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var _index = require("./path/index.js");
var _t = require("@babel/types");
var _context = require("./path/context.js");
const {
  VISITOR_KEYS
***REMOVED*** = _t;
class TraversalContext {
  constructor(scope, opts, state, parentPath) {
    this.queue = null;
    this.priorityQueue = null;
    this.parentPath = parentPath;
    this.scope = scope;
    this.state = state;
    this.opts = opts;
  ***REMOVED***
  shouldVisit(node) {
    const opts = this.opts;
    if (opts.enter || opts.exit) return true;
    if (opts[node.type]) return true;
    const keys = VISITOR_KEYS[node.type];
    if (!(keys != null && keys.length)) return false;
    for (const key of keys) {
      if (node[key]) {
        return true;
      ***REMOVED***
    ***REMOVED***
    return false;
  ***REMOVED***
  create(node, container, key, listKey) {
    return _index.default.get({
      parentPath: this.parentPath,
      parent: node,
      container,
      key: key,
      listKey
    ***REMOVED***);
  ***REMOVED***
  maybeQueue(path, notPriority) {
    if (this.queue) {
      if (notPriority) {
        this.queue.push(path);
      ***REMOVED*** else {
        this.priorityQueue.push(path);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  visitMultiple(container, parent, listKey) {
    if (container.length === 0) return false;
    const queue = [];
    for (let key = 0; key < container.length; key++) {
      const node = container[key];
      if (node && this.shouldVisit(node)) {
        queue.push(this.create(parent, container, key, listKey));
      ***REMOVED***
    ***REMOVED***
    return this.visitQueue(queue);
  ***REMOVED***
  visitSingle(node, key) {
    if (this.shouldVisit(node[key])) {
      return this.visitQueue([this.create(node, node, key)]);
    ***REMOVED*** else {
      return false;
    ***REMOVED***
  ***REMOVED***
  visitQueue(queue) {
    this.queue = queue;
    this.priorityQueue = [];
    const visited = new WeakSet();
    let stop = false;
    let visitIndex = 0;
    for (; visitIndex < queue.length;) {
      const path = queue[visitIndex];
      visitIndex++;
      _context.resync.call(path);
      if (path.contexts.length === 0 || path.contexts[path.contexts.length - 1] !== this) {
        _context.pushContext.call(path, this);
      ***REMOVED***
      if (path.key === null) continue;
      const {
        node
      ***REMOVED*** = path;
      if (visited.has(node)) continue;
      if (node) visited.add(node);
      if (path.visit()) {
        stop = true;
        break;
      ***REMOVED***
      if (this.priorityQueue.length) {
        stop = this.visitQueue(this.priorityQueue);
        this.priorityQueue = [];
        this.queue = queue;
        if (stop) break;
      ***REMOVED***
    ***REMOVED***
    for (let i = 0; i < visitIndex; i++) {
      _context.popContext.call(queue[i]);
    ***REMOVED***
    this.queue = null;
    return stop;
  ***REMOVED***
  visit(node, key) {
    const nodes = node[key];
    if (!nodes) return false;
    if (Array.isArray(nodes)) {
      return this.visitMultiple(nodes, node, key);
    ***REMOVED*** else {
      return this.visitSingle(node, key);
    ***REMOVED***
  ***REMOVED***
***REMOVED***
exports.default = TraversalContext;
