// eslint/lib/rules/accessor-pairs.js
/**
 * @fileoverview Rule to enforce getter and setter pairs in objects and classes.
 * @author Gyandeep Singh
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Typedefs
//------------------------------------------------------------------------------

/**
 * Property name if it can be computed statically, otherwise the list of the tokens of the key node.
 * @typedef {string|Token[]***REMOVED*** Key
 */

/**
 * Accessor nodes with the same key.
 * @typedef {Object***REMOVED*** AccessorData
 * @property {Key***REMOVED*** key Accessor's key
 * @property {ASTNode[]***REMOVED*** getters List of getter nodes.
 * @property {ASTNode[]***REMOVED*** setters List of setter nodes.
 */

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Checks whether or not the given lists represent the equal tokens in the same order.
 * Tokens are compared by their properties, not by instance.
 * @param {Token[]***REMOVED*** left First list of tokens.
 * @param {Token[]***REMOVED*** right Second list of tokens.
 * @returns {boolean***REMOVED*** `true` if the lists have same tokens.
 */
function areEqualTokenLists(left, right) {
    if (left.length !== right.length) {
        return false;
    ***REMOVED***

    for (let i = 0; i < left.length; i++) {
        const leftToken = left[i],
            rightToken = right[i];

        if (leftToken.type !== rightToken.type || leftToken.value !== rightToken.value) {
            return false;
        ***REMOVED***
    ***REMOVED***

    return true;
***REMOVED***

/**
 * Checks whether or not the given keys are equal.
 * @param {Key***REMOVED*** left First key.
 * @param {Key***REMOVED*** right Second key.
 * @returns {boolean***REMOVED*** `true` if the keys are equal.
 */
function areEqualKeys(left, right) {
    if (typeof left === "string" && typeof right === "string") {

        // Statically computed names.
        return left === right;
    ***REMOVED***
    if (Array.isArray(left) && Array.isArray(right)) {

        // Token lists.
        return areEqualTokenLists(left, right);
    ***REMOVED***

    return false;
***REMOVED***

/**
 * Checks whether or not a given node is of an accessor kind ('get' or 'set').
 * @param {ASTNode***REMOVED*** node A node to check.
 * @returns {boolean***REMOVED*** `true` if the node is of an accessor kind.
 */
function isAccessorKind(node) {
    return node.kind === "get" || node.kind === "set";
***REMOVED***

/**
 * Checks whether or not a given node is an argument of a specified method call.
 * @param {ASTNode***REMOVED*** node A node to check.
 * @param {number***REMOVED*** index An expected index of the node in arguments.
 * @param {string***REMOVED*** object An expected name of the object of the method.
 * @param {string***REMOVED*** property An expected name of the method.
 * @returns {boolean***REMOVED*** `true` if the node is an argument of the specified method call.
 */
function isArgumentOfMethodCall(node, index, object, property) {
    const parent = node.parent;

    return (
        parent.type === "CallExpression" &&
        astUtils.isSpecificMemberAccess(parent.callee, object, property) &&
        parent.arguments[index] === node
    );
***REMOVED***

/**
 * Checks whether or not a given node is a property descriptor.
 * @param {ASTNode***REMOVED*** node A node to check.
 * @returns {boolean***REMOVED*** `true` if the node is a property descriptor.
 */
function isPropertyDescriptor(node) {

    // Object.defineProperty(obj, "foo", {set: ...***REMOVED***)
    if (isArgumentOfMethodCall(node, 2, "Object", "defineProperty") ||
        isArgumentOfMethodCall(node, 2, "Reflect", "defineProperty")
    ) {
        return true;
    ***REMOVED***

    /*
     * Object.defineProperties(obj, {foo: {set: ...***REMOVED******REMOVED***)
     * Object.create(proto, {foo: {set: ...***REMOVED******REMOVED***)
     */
    const grandparent = node.parent.parent;

    return grandparent.type === "ObjectExpression" && (
        isArgumentOfMethodCall(grandparent, 1, "Object", "create") ||
        isArgumentOfMethodCall(grandparent, 1, "Object", "defineProperties")
    );
***REMOVED***

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "enforce getter and setter pairs in objects and classes",
            category: "Best Practices",
            recommended: false,
            url: "https://eslint.org/docs/rules/accessor-pairs"
        ***REMOVED***,

        schema: [{
            type: "object",
            properties: {
                getWithoutSet: {
                    type: "boolean",
                    default: false
                ***REMOVED***,
                setWithoutGet: {
                    type: "boolean",
                    default: true
                ***REMOVED***,
                enforceForClassMembers: {
                    type: "boolean",
                    default: true
                ***REMOVED***
            ***REMOVED***,
            additionalProperties: false
        ***REMOVED***],

        messages: {
            missingGetterInPropertyDescriptor: "Getter is not present in property descriptor.",
            missingSetterInPropertyDescriptor: "Setter is not present in property descriptor.",
            missingGetterInObjectLiteral: "Getter is not present for {{ name ***REMOVED******REMOVED***.",
            missingSetterInObjectLiteral: "Setter is not present for {{ name ***REMOVED******REMOVED***.",
            missingGetterInClass: "Getter is not present for class {{ name ***REMOVED******REMOVED***.",
            missingSetterInClass: "Setter is not present for class {{ name ***REMOVED******REMOVED***."
        ***REMOVED***
    ***REMOVED***,
    create(context) {
        const config = context.options[0] || {***REMOVED***;
        const checkGetWithoutSet = config.getWithoutSet === true;
        const checkSetWithoutGet = config.setWithoutGet !== false;
        const enforceForClassMembers = config.enforceForClassMembers !== false;
        const sourceCode = context.getSourceCode();

        /**
         * Reports the given node.
         * @param {ASTNode***REMOVED*** node The node to report.
         * @param {string***REMOVED*** messageKind "missingGetter" or "missingSetter".
         * @returns {void***REMOVED***
         * @private
         */
        function report(node, messageKind) {
            if (node.type === "Property") {
                context.report({
                    node,
                    messageId: `${messageKind***REMOVED***InObjectLiteral`,
                    loc: astUtils.getFunctionHeadLoc(node.value, sourceCode),
                    data: { name: astUtils.getFunctionNameWithKind(node.value) ***REMOVED***
                ***REMOVED***);
            ***REMOVED*** else if (node.type === "MethodDefinition") {
                context.report({
                    node,
                    messageId: `${messageKind***REMOVED***InClass`,
                    loc: astUtils.getFunctionHeadLoc(node.value, sourceCode),
                    data: { name: astUtils.getFunctionNameWithKind(node.value) ***REMOVED***
                ***REMOVED***);
            ***REMOVED*** else {
                context.report({
                    node,
                    messageId: `${messageKind***REMOVED***InPropertyDescriptor`
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        /**
         * Reports each of the nodes in the given list using the same messageId.
         * @param {ASTNode[]***REMOVED*** nodes Nodes to report.
         * @param {string***REMOVED*** messageKind "missingGetter" or "missingSetter".
         * @returns {void***REMOVED***
         * @private
         */
        function reportList(nodes, messageKind) {
            for (const node of nodes) {
                report(node, messageKind);
            ***REMOVED***
        ***REMOVED***

        /**
         * Creates a new `AccessorData` object for the given getter or setter node.
         * @param {ASTNode***REMOVED*** node A getter or setter node.
         * @returns {AccessorData***REMOVED*** New `AccessorData` object that contains the given node.
         * @private
         */
        function createAccessorData(node) {
            const name = astUtils.getStaticPropertyName(node);
            const key = (name !== null) ? name : sourceCode.getTokens(node.key);

            return {
                key,
                getters: node.kind === "get" ? [node] : [],
                setters: node.kind === "set" ? [node] : []
            ***REMOVED***;
        ***REMOVED***

        /**
         * Merges the given `AccessorData` object into the given accessors list.
         * @param {AccessorData[]***REMOVED*** accessors The list to merge into.
         * @param {AccessorData***REMOVED*** accessorData The object to merge.
         * @returns {AccessorData[]***REMOVED*** The same instance with the merged object.
         * @private
         */
        function mergeAccessorData(accessors, accessorData) {
            const equalKeyElement = accessors.find(a => areEqualKeys(a.key, accessorData.key));

            if (equalKeyElement) {
                equalKeyElement.getters.push(...accessorData.getters);
                equalKeyElement.setters.push(...accessorData.setters);
            ***REMOVED*** else {
                accessors.push(accessorData);
            ***REMOVED***

            return accessors;
        ***REMOVED***

        /**
         * Checks accessor pairs in the given list of nodes.
         * @param {ASTNode[]***REMOVED*** nodes The list to check.
         * @returns {void***REMOVED***
         * @private
         */
        function checkList(nodes) {
            const accessors = nodes
                .filter(isAccessorKind)
                .map(createAccessorData)
                .reduce(mergeAccessorData, []);

            for (const { getters, setters ***REMOVED*** of accessors) {
                if (checkSetWithoutGet && setters.length && !getters.length) {
                    reportList(setters, "missingGetter");
                ***REMOVED***
                if (checkGetWithoutSet && getters.length && !setters.length) {
                    reportList(getters, "missingSetter");
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        /**
         * Checks accessor pairs in an object literal.
         * @param {ASTNode***REMOVED*** node `ObjectExpression` node to check.
         * @returns {void***REMOVED***
         * @private
         */
        function checkObjectLiteral(node) {
            checkList(node.properties.filter(p => p.type === "Property"));
        ***REMOVED***

        /**
         * Checks accessor pairs in a property descriptor.
         * @param {ASTNode***REMOVED*** node Property descriptor `ObjectExpression` node to check.
         * @returns {void***REMOVED***
         * @private
         */
        function checkPropertyDescriptor(node) {
            const namesToCheck = node.properties
                .filter(p => p.type === "Property" && p.kind === "init" && !p.computed)
                .map(({ key ***REMOVED***) => key.name);

            const hasGetter = namesToCheck.includes("get");
            const hasSetter = namesToCheck.includes("set");

            if (checkSetWithoutGet && hasSetter && !hasGetter) {
                report(node, "missingGetter");
            ***REMOVED***
            if (checkGetWithoutSet && hasGetter && !hasSetter) {
                report(node, "missingSetter");
            ***REMOVED***
        ***REMOVED***

        /**
         * Checks the given object expression as an object literal and as a possible property descriptor.
         * @param {ASTNode***REMOVED*** node `ObjectExpression` node to check.
         * @returns {void***REMOVED***
         * @private
         */
        function checkObjectExpression(node) {
            checkObjectLiteral(node);
            if (isPropertyDescriptor(node)) {
                checkPropertyDescriptor(node);
            ***REMOVED***
        ***REMOVED***

        /**
         * Checks the given class body.
         * @param {ASTNode***REMOVED*** node `ClassBody` node to check.
         * @returns {void***REMOVED***
         * @private
         */
        function checkClassBody(node) {
            const methodDefinitions = node.body.filter(m => m.type === "MethodDefinition");

            checkList(methodDefinitions.filter(m => m.static));
            checkList(methodDefinitions.filter(m => !m.static));
        ***REMOVED***

        const listeners = {***REMOVED***;

        if (checkSetWithoutGet || checkGetWithoutSet) {
            listeners.ObjectExpression = checkObjectExpression;
            if (enforceForClassMembers) {
                listeners.ClassBody = checkClassBody;
            ***REMOVED***
        ***REMOVED***

        return listeners;
    ***REMOVED***
***REMOVED***;
