// eslint/lib/rules/comma-spacing.js
/**
 * @fileoverview Comma spacing - validates spacing before and after comma
 * @author Vignesh Anand aka vegetableman.
 */
"use strict";

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "layout",

        docs: {
            description: "enforce consistent spacing before and after commas",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/comma-spacing"
        ***REMOVED***,

        fixable: "whitespace",

        schema: [
            {
                type: "object",
                properties: {
                    before: {
                        type: "boolean",
                        default: false
                    ***REMOVED***,
                    after: {
                        type: "boolean",
                        default: true
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],

        messages: {
            missing: "A space is required {{loc***REMOVED******REMOVED*** ','.",
            unexpected: "There should be no space {{loc***REMOVED******REMOVED*** ','."
        ***REMOVED***
    ***REMOVED***,

    create(context) {

        const sourceCode = context.getSourceCode();
        const tokensAndComments = sourceCode.tokensAndComments;

        const options = {
            before: context.options[0] ? context.options[0].before : false,
            after: context.options[0] ? context.options[0].after : true
        ***REMOVED***;

        //--------------------------------------------------------------------------
        // Helpers
        //--------------------------------------------------------------------------

        // list of comma tokens to ignore for the check of leading whitespace
        const commaTokensToIgnore = [];

        /**
         * Reports a spacing error with an appropriate message.
         * @param {ASTNode***REMOVED*** node The binary expression node to report.
         * @param {string***REMOVED*** loc Is the error "before" or "after" the comma?
         * @param {ASTNode***REMOVED*** otherNode The node at the left or right of `node`
         * @returns {void***REMOVED***
         * @private
         */
        function report(node, loc, otherNode) {
            context.report({
                node,
                fix(fixer) {
                    if (options[loc]) {
                        if (loc === "before") {
                            return fixer.insertTextBefore(node, " ");
                        ***REMOVED***
                        return fixer.insertTextAfter(node, " ");

                    ***REMOVED***
                    let start, end;
                    const newText = "";

                    if (loc === "before") {
                        start = otherNode.range[1];
                        end = node.range[0];
                    ***REMOVED*** else {
                        start = node.range[1];
                        end = otherNode.range[0];
                    ***REMOVED***

                    return fixer.replaceTextRange([start, end], newText);

                ***REMOVED***,
                messageId: options[loc] ? "missing" : "unexpected",
                data: {
                    loc
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Validates the spacing around a comma token.
         * @param {Object***REMOVED*** tokens The tokens to be validated.
         * @param {Token***REMOVED*** tokens.comma The token representing the comma.
         * @param {Token***REMOVED*** [tokens.left] The last token before the comma.
         * @param {Token***REMOVED*** [tokens.right] The first token after the comma.
         * @param {Token|ASTNode***REMOVED*** reportItem The item to use when reporting an error.
         * @returns {void***REMOVED***
         * @private
         */
        function validateCommaItemSpacing(tokens, reportItem) {
            if (tokens.left && astUtils.isTokenOnSameLine(tokens.left, tokens.comma) &&
                    (options.before !== sourceCode.isSpaceBetweenTokens(tokens.left, tokens.comma))
            ) {
                report(reportItem, "before", tokens.left);
            ***REMOVED***

            if (tokens.right && astUtils.isClosingParenToken(tokens.right)) {
                return;
            ***REMOVED***

            if (tokens.right && !options.after && tokens.right.type === "Line") {
                return;
            ***REMOVED***

            if (tokens.right && astUtils.isTokenOnSameLine(tokens.comma, tokens.right) &&
                    (options.after !== sourceCode.isSpaceBetweenTokens(tokens.comma, tokens.right))
            ) {
                report(reportItem, "after", tokens.right);
            ***REMOVED***
        ***REMOVED***

        /**
         * Adds null elements of the given ArrayExpression or ArrayPattern node to the ignore list.
         * @param {ASTNode***REMOVED*** node An ArrayExpression or ArrayPattern node.
         * @returns {void***REMOVED***
         */
        function addNullElementsToIgnoreList(node) {
            let previousToken = sourceCode.getFirstToken(node);

            node.elements.forEach(element => {
                let token;

                if (element === null) {
                    token = sourceCode.getTokenAfter(previousToken);

                    if (astUtils.isCommaToken(token)) {
                        commaTokensToIgnore.push(token);
                    ***REMOVED***
                ***REMOVED*** else {
                    token = sourceCode.getTokenAfter(element);
                ***REMOVED***

                previousToken = token;
            ***REMOVED***);
        ***REMOVED***

        //--------------------------------------------------------------------------
        // Public
        //--------------------------------------------------------------------------

        return {
            "Program:exit"() {
                tokensAndComments.forEach((token, i) => {

                    if (!astUtils.isCommaToken(token)) {
                        return;
                    ***REMOVED***

                    if (token && token.type === "JSXText") {
                        return;
                    ***REMOVED***

                    const previousToken = tokensAndComments[i - 1];
                    const nextToken = tokensAndComments[i + 1];

                    validateCommaItemSpacing({
                        comma: token,
                        left: astUtils.isCommaToken(previousToken) || commaTokensToIgnore.includes(token) ? null : previousToken,
                        right: astUtils.isCommaToken(nextToken) ? null : nextToken
                    ***REMOVED***, token);
                ***REMOVED***);
            ***REMOVED***,
            ArrayExpression: addNullElementsToIgnoreList,
            ArrayPattern: addNullElementsToIgnoreList

        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
