// eslint/lib/rules/computed-property-spacing.js
/**
 * @fileoverview Disallows or enforces spaces inside computed properties.
 * @author Jamund Ferguson
 * @deprecated in ESLint v8.53.0
 */
"use strict";

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

/** @type {import('../shared/types').Rule***REMOVED*** */
module.exports = {
    meta: {
        deprecated: true,
        replacedBy: [],
        type: "layout",

        docs: {
            description: "Enforce consistent spacing inside computed property brackets",
            recommended: false,
            url: "https://eslint.org/docs/latest/rules/computed-property-spacing"
        ***REMOVED***,

        fixable: "whitespace",

        schema: [
            {
                enum: ["always", "never"]
            ***REMOVED***,
            {
                type: "object",
                properties: {
                    enforceForClassMembers: {
                        type: "boolean",
                        default: true
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],

        messages: {
            unexpectedSpaceBefore: "There should be no space before '{{tokenValue***REMOVED******REMOVED***'.",
            unexpectedSpaceAfter: "There should be no space after '{{tokenValue***REMOVED******REMOVED***'.",

            missingSpaceBefore: "A space is required before '{{tokenValue***REMOVED******REMOVED***'.",
            missingSpaceAfter: "A space is required after '{{tokenValue***REMOVED******REMOVED***'."
        ***REMOVED***
    ***REMOVED***,

    create(context) {
        const sourceCode = context.sourceCode;
        const propertyNameMustBeSpaced = context.options[0] === "always"; // default is "never"
        const enforceForClassMembers = !context.options[1] || context.options[1].enforceForClassMembers;

        //--------------------------------------------------------------------------
        // Helpers
        //--------------------------------------------------------------------------

        /**
         * Reports that there shouldn't be a space after the first token
         * @param {ASTNode***REMOVED*** node The node to report in the event of an error.
         * @param {Token***REMOVED*** token The token to use for the report.
         * @param {Token***REMOVED*** tokenAfter The token after `token`.
         * @returns {void***REMOVED***
         */
        function reportNoBeginningSpace(node, token, tokenAfter) {
            context.report({
                node,
                loc: { start: token.loc.end, end: tokenAfter.loc.start ***REMOVED***,
                messageId: "unexpectedSpaceAfter",
                data: {
                    tokenValue: token.value
                ***REMOVED***,
                fix(fixer) {
                    return fixer.removeRange([token.range[1], tokenAfter.range[0]]);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Reports that there shouldn't be a space before the last token
         * @param {ASTNode***REMOVED*** node The node to report in the event of an error.
         * @param {Token***REMOVED*** token The token to use for the report.
         * @param {Token***REMOVED*** tokenBefore The token before `token`.
         * @returns {void***REMOVED***
         */
        function reportNoEndingSpace(node, token, tokenBefore) {
            context.report({
                node,
                loc: { start: tokenBefore.loc.end, end: token.loc.start ***REMOVED***,
                messageId: "unexpectedSpaceBefore",
                data: {
                    tokenValue: token.value
                ***REMOVED***,
                fix(fixer) {
                    return fixer.removeRange([tokenBefore.range[1], token.range[0]]);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Reports that there should be a space after the first token
         * @param {ASTNode***REMOVED*** node The node to report in the event of an error.
         * @param {Token***REMOVED*** token The token to use for the report.
         * @returns {void***REMOVED***
         */
        function reportRequiredBeginningSpace(node, token) {
            context.report({
                node,
                loc: token.loc,
                messageId: "missingSpaceAfter",
                data: {
                    tokenValue: token.value
                ***REMOVED***,
                fix(fixer) {
                    return fixer.insertTextAfter(token, " ");
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Reports that there should be a space before the last token
         * @param {ASTNode***REMOVED*** node The node to report in the event of an error.
         * @param {Token***REMOVED*** token The token to use for the report.
         * @returns {void***REMOVED***
         */
        function reportRequiredEndingSpace(node, token) {
            context.report({
                node,
                loc: token.loc,
                messageId: "missingSpaceBefore",
                data: {
                    tokenValue: token.value
                ***REMOVED***,
                fix(fixer) {
                    return fixer.insertTextBefore(token, " ");
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Returns a function that checks the spacing of a node on the property name
         * that was passed in.
         * @param {string***REMOVED*** propertyName The property on the node to check for spacing
         * @returns {Function***REMOVED*** A function that will check spacing on a node
         */
        function checkSpacing(propertyName) {
            return function(node) {
                if (!node.computed) {
                    return;
                ***REMOVED***

                const property = node[propertyName];

                const before = sourceCode.getTokenBefore(property, astUtils.isOpeningBracketToken),
                    first = sourceCode.getTokenAfter(before, { includeComments: true ***REMOVED***),
                    after = sourceCode.getTokenAfter(property, astUtils.isClosingBracketToken),
                    last = sourceCode.getTokenBefore(after, { includeComments: true ***REMOVED***);

                if (astUtils.isTokenOnSameLine(before, first)) {
                    if (propertyNameMustBeSpaced) {
                        if (!sourceCode.isSpaceBetweenTokens(before, first) && astUtils.isTokenOnSameLine(before, first)) {
                            reportRequiredBeginningSpace(node, before);
                        ***REMOVED***
                    ***REMOVED*** else {
                        if (sourceCode.isSpaceBetweenTokens(before, first)) {
                            reportNoBeginningSpace(node, before, first);
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***

                if (astUtils.isTokenOnSameLine(last, after)) {
                    if (propertyNameMustBeSpaced) {
                        if (!sourceCode.isSpaceBetweenTokens(last, after) && astUtils.isTokenOnSameLine(last, after)) {
                            reportRequiredEndingSpace(node, after);
                        ***REMOVED***
                    ***REMOVED*** else {
                        if (sourceCode.isSpaceBetweenTokens(last, after)) {
                            reportNoEndingSpace(node, after, last);
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***;
        ***REMOVED***


        //--------------------------------------------------------------------------
        // Public
        //--------------------------------------------------------------------------

        const listeners = {
            Property: checkSpacing("key"),
            MemberExpression: checkSpacing("property")
        ***REMOVED***;

        if (enforceForClassMembers) {
            listeners.MethodDefinition =
                listeners.PropertyDefinition = listeners.Property;
        ***REMOVED***

        return listeners;

    ***REMOVED***
***REMOVED***;
