// eslint/lib/rules/default-param-last.js
/**
 * @fileoverview enforce default parameters to be last
 * @author Chiawen Chen
 */

"use strict";

module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "enforce default parameters to be last",
            category: "Best Practices",
            recommended: false,
            url: "https://eslint.org/docs/rules/default-param-last"
        ***REMOVED***,

        schema: [],

        messages: {
            shouldBeLast: "Default parameters should be last."
        ***REMOVED***
    ***REMOVED***,

    create(context) {

        // eslint-disable-next-line jsdoc/require-description
        /**
         * @param {ASTNode***REMOVED*** node function node
         * @returns {void***REMOVED***
         */
        function handleFunction(node) {
            let hasSeenPlainParam = false;

            for (let i = node.params.length - 1; i >= 0; i -= 1) {
                const param = node.params[i];

                if (
                    param.type !== "AssignmentPattern" &&
                    param.type !== "RestElement"
                ) {
                    hasSeenPlainParam = true;
                    continue;
                ***REMOVED***

                if (hasSeenPlainParam && param.type === "AssignmentPattern") {
                    context.report({
                        node: param,
                        messageId: "shouldBeLast"
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        return {
            FunctionDeclaration: handleFunction,
            FunctionExpression: handleFunction,
            ArrowFunctionExpression: handleFunction
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
