// eslint/lib/rules/func-style.js
/**
 * @fileoverview Rule to enforce a particular function style
 * @author Nicholas C. Zakas
 */
"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "enforce the consistent use of either `function` declarations or expressions",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/func-style"
        ***REMOVED***,

        schema: [
            {
                enum: ["declaration", "expression"]
            ***REMOVED***,
            {
                type: "object",
                properties: {
                    allowArrowFunctions: {
                        type: "boolean",
                        default: false
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],

        messages: {
            expression: "Expected a function expression.",
            declaration: "Expected a function declaration."
        ***REMOVED***
    ***REMOVED***,

    create(context) {

        const style = context.options[0],
            allowArrowFunctions = context.options[1] && context.options[1].allowArrowFunctions,
            enforceDeclarations = (style === "declaration"),
            stack = [];

        const nodesToCheck = {
            FunctionDeclaration(node) {
                stack.push(false);

                if (!enforceDeclarations && node.parent.type !== "ExportDefaultDeclaration") {
                    context.report({ node, messageId: "expression" ***REMOVED***);
                ***REMOVED***
            ***REMOVED***,
            "FunctionDeclaration:exit"() {
                stack.pop();
            ***REMOVED***,

            FunctionExpression(node) {
                stack.push(false);

                if (enforceDeclarations && node.parent.type === "VariableDeclarator") {
                    context.report({ node: node.parent, messageId: "declaration" ***REMOVED***);
                ***REMOVED***
            ***REMOVED***,
            "FunctionExpression:exit"() {
                stack.pop();
            ***REMOVED***,

            ThisExpression() {
                if (stack.length > 0) {
                    stack[stack.length - 1] = true;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;

        if (!allowArrowFunctions) {
            nodesToCheck.ArrowFunctionExpression = function() {
                stack.push(false);
            ***REMOVED***;

            nodesToCheck["ArrowFunctionExpression:exit"] = function(node) {
                const hasThisExpr = stack.pop();

                if (enforceDeclarations && !hasThisExpr && node.parent.type === "VariableDeclarator") {
                    context.report({ node: node.parent, messageId: "declaration" ***REMOVED***);
                ***REMOVED***
            ***REMOVED***;
        ***REMOVED***

        return nodesToCheck;

    ***REMOVED***
***REMOVED***;
