// eslint/lib/rules/grouped-accessor-pairs.js
/**
 * @fileoverview Rule to require grouped accessor pairs in object literals and classes
 * @author Milos Djermanovic
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Typedefs
//------------------------------------------------------------------------------

/**
 * Property name if it can be computed statically, otherwise the list of the tokens of the key node.
 * @typedef {string|Token[]***REMOVED*** Key
 */

/**
 * Accessor nodes with the same key.
 * @typedef {Object***REMOVED*** AccessorData
 * @property {Key***REMOVED*** key Accessor's key
 * @property {ASTNode[]***REMOVED*** getters List of getter nodes.
 * @property {ASTNode[]***REMOVED*** setters List of setter nodes.
 */

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Checks whether or not the given lists represent the equal tokens in the same order.
 * Tokens are compared by their properties, not by instance.
 * @param {Token[]***REMOVED*** left First list of tokens.
 * @param {Token[]***REMOVED*** right Second list of tokens.
 * @returns {boolean***REMOVED*** `true` if the lists have same tokens.
 */
function areEqualTokenLists(left, right) {
    if (left.length !== right.length) {
        return false;
    ***REMOVED***

    for (let i = 0; i < left.length; i++) {
        const leftToken = left[i],
            rightToken = right[i];

        if (leftToken.type !== rightToken.type || leftToken.value !== rightToken.value) {
            return false;
        ***REMOVED***
    ***REMOVED***

    return true;
***REMOVED***

/**
 * Checks whether or not the given keys are equal.
 * @param {Key***REMOVED*** left First key.
 * @param {Key***REMOVED*** right Second key.
 * @returns {boolean***REMOVED*** `true` if the keys are equal.
 */
function areEqualKeys(left, right) {
    if (typeof left === "string" && typeof right === "string") {

        // Statically computed names.
        return left === right;
    ***REMOVED***
    if (Array.isArray(left) && Array.isArray(right)) {

        // Token lists.
        return areEqualTokenLists(left, right);
    ***REMOVED***

    return false;
***REMOVED***

/**
 * Checks whether or not a given node is of an accessor kind ('get' or 'set').
 * @param {ASTNode***REMOVED*** node A node to check.
 * @returns {boolean***REMOVED*** `true` if the node is of an accessor kind.
 */
function isAccessorKind(node) {
    return node.kind === "get" || node.kind === "set";
***REMOVED***

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "require grouped accessor pairs in object literals and classes",
            category: "Best Practices",
            recommended: false,
            url: "https://eslint.org/docs/rules/grouped-accessor-pairs"
        ***REMOVED***,

        schema: [
            {
                enum: ["anyOrder", "getBeforeSet", "setBeforeGet"]
            ***REMOVED***
        ],

        messages: {
            notGrouped: "Accessor pair {{ formerName ***REMOVED******REMOVED*** and {{ latterName ***REMOVED******REMOVED*** should be grouped.",
            invalidOrder: "Expected {{ latterName ***REMOVED******REMOVED*** to be before {{ formerName ***REMOVED******REMOVED***."
        ***REMOVED***
    ***REMOVED***,

    create(context) {
        const order = context.options[0] || "anyOrder";
        const sourceCode = context.getSourceCode();

        /**
         * Reports the given accessor pair.
         * @param {string***REMOVED*** messageId messageId to report.
         * @param {ASTNode***REMOVED*** formerNode getter/setter node that is defined before `latterNode`.
         * @param {ASTNode***REMOVED*** latterNode getter/setter node that is defined after `formerNode`.
         * @returns {void***REMOVED***
         * @private
         */
        function report(messageId, formerNode, latterNode) {
            context.report({
                node: latterNode,
                messageId,
                loc: astUtils.getFunctionHeadLoc(latterNode.value, sourceCode),
                data: {
                    formerName: astUtils.getFunctionNameWithKind(formerNode.value),
                    latterName: astUtils.getFunctionNameWithKind(latterNode.value)
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Creates a new `AccessorData` object for the given getter or setter node.
         * @param {ASTNode***REMOVED*** node A getter or setter node.
         * @returns {AccessorData***REMOVED*** New `AccessorData` object that contains the given node.
         * @private
         */
        function createAccessorData(node) {
            const name = astUtils.getStaticPropertyName(node);
            const key = (name !== null) ? name : sourceCode.getTokens(node.key);

            return {
                key,
                getters: node.kind === "get" ? [node] : [],
                setters: node.kind === "set" ? [node] : []
            ***REMOVED***;
        ***REMOVED***

        /**
         * Merges the given `AccessorData` object into the given accessors list.
         * @param {AccessorData[]***REMOVED*** accessors The list to merge into.
         * @param {AccessorData***REMOVED*** accessorData The object to merge.
         * @returns {AccessorData[]***REMOVED*** The same instance with the merged object.
         * @private
         */
        function mergeAccessorData(accessors, accessorData) {
            const equalKeyElement = accessors.find(a => areEqualKeys(a.key, accessorData.key));

            if (equalKeyElement) {
                equalKeyElement.getters.push(...accessorData.getters);
                equalKeyElement.setters.push(...accessorData.setters);
            ***REMOVED*** else {
                accessors.push(accessorData);
            ***REMOVED***

            return accessors;
        ***REMOVED***

        /**
         * Checks accessor pairs in the given list of nodes.
         * @param {ASTNode[]***REMOVED*** nodes The list to check.
         * @param {Function***REMOVED*** shouldCheck – Predicate that returns `true` if the node should be checked.
         * @returns {void***REMOVED***
         * @private
         */
        function checkList(nodes, shouldCheck) {
            const accessors = nodes
                .filter(shouldCheck)
                .filter(isAccessorKind)
                .map(createAccessorData)
                .reduce(mergeAccessorData, []);

            for (const { getters, setters ***REMOVED*** of accessors) {

                // Don't report accessor properties that have duplicate getters or setters.
                if (getters.length === 1 && setters.length === 1) {
                    const [getter] = getters,
                        [setter] = setters,
                        getterIndex = nodes.indexOf(getter),
                        setterIndex = nodes.indexOf(setter),
                        formerNode = getterIndex < setterIndex ? getter : setter,
                        latterNode = getterIndex < setterIndex ? setter : getter;

                    if (Math.abs(getterIndex - setterIndex) > 1) {
                        report("notGrouped", formerNode, latterNode);
                    ***REMOVED*** else if (
                        (order === "getBeforeSet" && getterIndex > setterIndex) ||
                        (order === "setBeforeGet" && getterIndex < setterIndex)
                    ) {
                        report("invalidOrder", formerNode, latterNode);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        return {
            ObjectExpression(node) {
                checkList(node.properties, n => n.type === "Property");
            ***REMOVED***,
            ClassBody(node) {
                checkList(node.body, n => n.type === "MethodDefinition" && !n.static);
                checkList(node.body, n => n.type === "MethodDefinition" && n.static);
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
