// eslint/lib/rules/handle-callback-err.js
/**
 * @fileoverview Ensure handling of errors when we know they exist.
 * @author Jamund Ferguson
 * @deprecated in ESLint v7.0.0
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

/** @type {import('../shared/types').Rule***REMOVED*** */
module.exports = {
    meta: {
        deprecated: true,

        replacedBy: [],

        type: "suggestion",

        docs: {
            description: "Require error handling in callbacks",
            recommended: false,
            url: "https://eslint.org/docs/latest/rules/handle-callback-err"
        ***REMOVED***,

        schema: [
            {
                type: "string"
            ***REMOVED***
        ],
        messages: {
            expected: "Expected error to be handled."
        ***REMOVED***
    ***REMOVED***,

    create(context) {

        const errorArgument = context.options[0] || "err";
        const sourceCode = context.sourceCode;

        /**
         * Checks if the given argument should be interpreted as a regexp pattern.
         * @param {string***REMOVED*** stringToCheck The string which should be checked.
         * @returns {boolean***REMOVED*** Whether or not the string should be interpreted as a pattern.
         */
        function isPattern(stringToCheck) {
            const firstChar = stringToCheck[0];

            return firstChar === "^";
        ***REMOVED***

        /**
         * Checks if the given name matches the configured error argument.
         * @param {string***REMOVED*** name The name which should be compared.
         * @returns {boolean***REMOVED*** Whether or not the given name matches the configured error variable name.
         */
        function matchesConfiguredErrorName(name) {
            if (isPattern(errorArgument)) {
                const regexp = new RegExp(errorArgument, "u");

                return regexp.test(name);
            ***REMOVED***
            return name === errorArgument;
        ***REMOVED***

        /**
         * Get the parameters of a given function scope.
         * @param {Object***REMOVED*** scope The function scope.
         * @returns {Array***REMOVED*** All parameters of the given scope.
         */
        function getParameters(scope) {
            return scope.variables.filter(variable => variable.defs[0] && variable.defs[0].type === "Parameter");
        ***REMOVED***

        /**
         * Check to see if we're handling the error object properly.
         * @param {ASTNode***REMOVED*** node The AST node to check.
         * @returns {void***REMOVED***
         */
        function checkForError(node) {
            const scope = sourceCode.getScope(node),
                parameters = getParameters(scope),
                firstParameter = parameters[0];

            if (firstParameter && matchesConfiguredErrorName(firstParameter.name)) {
                if (firstParameter.references.length === 0) {
                    context.report({ node, messageId: "expected" ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        return {
            FunctionDeclaration: checkForError,
            FunctionExpression: checkForError,
            ArrowFunctionExpression: checkForError
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
