// eslint/lib/rules/lines-around-comment.js
/**
 * @fileoverview Enforces empty lines around comments.
 * @author Jamund Ferguson
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Return an array with with any line numbers that are empty.
 * @param {Array***REMOVED*** lines An array of each line of the file.
 * @returns {Array***REMOVED*** An array of line numbers.
 */
function getEmptyLineNums(lines) {
    const emptyLines = lines.map((line, i) => ({
        code: line.trim(),
        num: i + 1
    ***REMOVED***)).filter(line => !line.code).map(line => line.num);

    return emptyLines;
***REMOVED***

/**
 * Return an array with with any line numbers that contain comments.
 * @param {Array***REMOVED*** comments An array of comment tokens.
 * @returns {Array***REMOVED*** An array of line numbers.
 */
function getCommentLineNums(comments) {
    const lines = [];

    comments.forEach(token => {
        const start = token.loc.start.line;
        const end = token.loc.end.line;

        lines.push(start, end);
    ***REMOVED***);
    return lines;
***REMOVED***

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "layout",

        docs: {
            description: "require empty lines around comments",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/lines-around-comment"
        ***REMOVED***,

        fixable: "whitespace",

        schema: [
            {
                type: "object",
                properties: {
                    beforeBlockComment: {
                        type: "boolean",
                        default: true
                    ***REMOVED***,
                    afterBlockComment: {
                        type: "boolean",
                        default: false
                    ***REMOVED***,
                    beforeLineComment: {
                        type: "boolean",
                        default: false
                    ***REMOVED***,
                    afterLineComment: {
                        type: "boolean",
                        default: false
                    ***REMOVED***,
                    allowBlockStart: {
                        type: "boolean",
                        default: false
                    ***REMOVED***,
                    allowBlockEnd: {
                        type: "boolean",
                        default: false
                    ***REMOVED***,
                    allowClassStart: {
                        type: "boolean"
                    ***REMOVED***,
                    allowClassEnd: {
                        type: "boolean"
                    ***REMOVED***,
                    allowObjectStart: {
                        type: "boolean"
                    ***REMOVED***,
                    allowObjectEnd: {
                        type: "boolean"
                    ***REMOVED***,
                    allowArrayStart: {
                        type: "boolean"
                    ***REMOVED***,
                    allowArrayEnd: {
                        type: "boolean"
                    ***REMOVED***,
                    ignorePattern: {
                        type: "string"
                    ***REMOVED***,
                    applyDefaultIgnorePatterns: {
                        type: "boolean"
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],
        messages: {
            after: "Expected line after comment.",
            before: "Expected line before comment."
        ***REMOVED***
    ***REMOVED***,

    create(context) {

        const options = Object.assign({***REMOVED***, context.options[0]);
        const ignorePattern = options.ignorePattern;
        const defaultIgnoreRegExp = astUtils.COMMENTS_IGNORE_PATTERN;
        const customIgnoreRegExp = new RegExp(ignorePattern, "u");
        const applyDefaultIgnorePatterns = options.applyDefaultIgnorePatterns !== false;

        options.beforeBlockComment = typeof options.beforeBlockComment !== "undefined" ? options.beforeBlockComment : true;

        const sourceCode = context.getSourceCode();

        const lines = sourceCode.lines,
            numLines = lines.length + 1,
            comments = sourceCode.getAllComments(),
            commentLines = getCommentLineNums(comments),
            emptyLines = getEmptyLineNums(lines),
            commentAndEmptyLines = commentLines.concat(emptyLines);

        /**
         * Returns whether or not comments are on lines starting with or ending with code
         * @param {token***REMOVED*** token The comment token to check.
         * @returns {boolean***REMOVED*** True if the comment is not alone.
         */
        function codeAroundComment(token) {
            let currentToken = token;

            do {
                currentToken = sourceCode.getTokenBefore(currentToken, { includeComments: true ***REMOVED***);
            ***REMOVED*** while (currentToken && astUtils.isCommentToken(currentToken));

            if (currentToken && astUtils.isTokenOnSameLine(currentToken, token)) {
                return true;
            ***REMOVED***

            currentToken = token;
            do {
                currentToken = sourceCode.getTokenAfter(currentToken, { includeComments: true ***REMOVED***);
            ***REMOVED*** while (currentToken && astUtils.isCommentToken(currentToken));

            if (currentToken && astUtils.isTokenOnSameLine(token, currentToken)) {
                return true;
            ***REMOVED***

            return false;
        ***REMOVED***

        /**
         * Returns whether or not comments are inside a node type or not.
         * @param {ASTNode***REMOVED*** parent The Comment parent node.
         * @param {string***REMOVED*** nodeType The parent type to check against.
         * @returns {boolean***REMOVED*** True if the comment is inside nodeType.
         */
        function isParentNodeType(parent, nodeType) {
            return parent.type === nodeType ||
                (parent.body && parent.body.type === nodeType) ||
                (parent.consequent && parent.consequent.type === nodeType);
        ***REMOVED***

        /**
         * Returns the parent node that contains the given token.
         * @param {token***REMOVED*** token The token to check.
         * @returns {ASTNode***REMOVED*** The parent node that contains the given token.
         */
        function getParentNodeOfToken(token) {
            return sourceCode.getNodeByRangeIndex(token.range[0]);
        ***REMOVED***

        /**
         * Returns whether or not comments are at the parent start or not.
         * @param {token***REMOVED*** token The Comment token.
         * @param {string***REMOVED*** nodeType The parent type to check against.
         * @returns {boolean***REMOVED*** True if the comment is at parent start.
         */
        function isCommentAtParentStart(token, nodeType) {
            const parent = getParentNodeOfToken(token);

            return parent && isParentNodeType(parent, nodeType) &&
                    token.loc.start.line - parent.loc.start.line === 1;
        ***REMOVED***

        /**
         * Returns whether or not comments are at the parent end or not.
         * @param {token***REMOVED*** token The Comment token.
         * @param {string***REMOVED*** nodeType The parent type to check against.
         * @returns {boolean***REMOVED*** True if the comment is at parent end.
         */
        function isCommentAtParentEnd(token, nodeType) {
            const parent = getParentNodeOfToken(token);

            return parent && isParentNodeType(parent, nodeType) &&
                    parent.loc.end.line - token.loc.end.line === 1;
        ***REMOVED***

        /**
         * Returns whether or not comments are at the block start or not.
         * @param {token***REMOVED*** token The Comment token.
         * @returns {boolean***REMOVED*** True if the comment is at block start.
         */
        function isCommentAtBlockStart(token) {
            return isCommentAtParentStart(token, "ClassBody") || isCommentAtParentStart(token, "BlockStatement") || isCommentAtParentStart(token, "SwitchCase");
        ***REMOVED***

        /**
         * Returns whether or not comments are at the block end or not.
         * @param {token***REMOVED*** token The Comment token.
         * @returns {boolean***REMOVED*** True if the comment is at block end.
         */
        function isCommentAtBlockEnd(token) {
            return isCommentAtParentEnd(token, "ClassBody") || isCommentAtParentEnd(token, "BlockStatement") || isCommentAtParentEnd(token, "SwitchCase") || isCommentAtParentEnd(token, "SwitchStatement");
        ***REMOVED***

        /**
         * Returns whether or not comments are at the class start or not.
         * @param {token***REMOVED*** token The Comment token.
         * @returns {boolean***REMOVED*** True if the comment is at class start.
         */
        function isCommentAtClassStart(token) {
            return isCommentAtParentStart(token, "ClassBody");
        ***REMOVED***

        /**
         * Returns whether or not comments are at the class end or not.
         * @param {token***REMOVED*** token The Comment token.
         * @returns {boolean***REMOVED*** True if the comment is at class end.
         */
        function isCommentAtClassEnd(token) {
            return isCommentAtParentEnd(token, "ClassBody");
        ***REMOVED***

        /**
         * Returns whether or not comments are at the object start or not.
         * @param {token***REMOVED*** token The Comment token.
         * @returns {boolean***REMOVED*** True if the comment is at object start.
         */
        function isCommentAtObjectStart(token) {
            return isCommentAtParentStart(token, "ObjectExpression") || isCommentAtParentStart(token, "ObjectPattern");
        ***REMOVED***

        /**
         * Returns whether or not comments are at the object end or not.
         * @param {token***REMOVED*** token The Comment token.
         * @returns {boolean***REMOVED*** True if the comment is at object end.
         */
        function isCommentAtObjectEnd(token) {
            return isCommentAtParentEnd(token, "ObjectExpression") || isCommentAtParentEnd(token, "ObjectPattern");
        ***REMOVED***

        /**
         * Returns whether or not comments are at the array start or not.
         * @param {token***REMOVED*** token The Comment token.
         * @returns {boolean***REMOVED*** True if the comment is at array start.
         */
        function isCommentAtArrayStart(token) {
            return isCommentAtParentStart(token, "ArrayExpression") || isCommentAtParentStart(token, "ArrayPattern");
        ***REMOVED***

        /**
         * Returns whether or not comments are at the array end or not.
         * @param {token***REMOVED*** token The Comment token.
         * @returns {boolean***REMOVED*** True if the comment is at array end.
         */
        function isCommentAtArrayEnd(token) {
            return isCommentAtParentEnd(token, "ArrayExpression") || isCommentAtParentEnd(token, "ArrayPattern");
        ***REMOVED***

        /**
         * Checks if a comment token has lines around it (ignores inline comments)
         * @param {token***REMOVED*** token The Comment token.
         * @param {Object***REMOVED*** opts Options to determine the newline.
         * @param {boolean***REMOVED*** opts.after Should have a newline after this line.
         * @param {boolean***REMOVED*** opts.before Should have a newline before this line.
         * @returns {void***REMOVED***
         */
        function checkForEmptyLine(token, opts) {
            if (applyDefaultIgnorePatterns && defaultIgnoreRegExp.test(token.value)) {
                return;
            ***REMOVED***

            if (ignorePattern && customIgnoreRegExp.test(token.value)) {
                return;
            ***REMOVED***

            let after = opts.after,
                before = opts.before;

            const prevLineNum = token.loc.start.line - 1,
                nextLineNum = token.loc.end.line + 1,
                commentIsNotAlone = codeAroundComment(token);

            const blockStartAllowed = options.allowBlockStart &&
                    isCommentAtBlockStart(token) &&
                    !(options.allowClassStart === false &&
                    isCommentAtClassStart(token)),
                blockEndAllowed = options.allowBlockEnd && isCommentAtBlockEnd(token) && !(options.allowClassEnd === false && isCommentAtClassEnd(token)),
                classStartAllowed = options.allowClassStart && isCommentAtClassStart(token),
                classEndAllowed = options.allowClassEnd && isCommentAtClassEnd(token),
                objectStartAllowed = options.allowObjectStart && isCommentAtObjectStart(token),
                objectEndAllowed = options.allowObjectEnd && isCommentAtObjectEnd(token),
                arrayStartAllowed = options.allowArrayStart && isCommentAtArrayStart(token),
                arrayEndAllowed = options.allowArrayEnd && isCommentAtArrayEnd(token);

            const exceptionStartAllowed = blockStartAllowed || classStartAllowed || objectStartAllowed || arrayStartAllowed;
            const exceptionEndAllowed = blockEndAllowed || classEndAllowed || objectEndAllowed || arrayEndAllowed;

            // ignore top of the file and bottom of the file
            if (prevLineNum < 1) {
                before = false;
            ***REMOVED***
            if (nextLineNum >= numLines) {
                after = false;
            ***REMOVED***

            // we ignore all inline comments
            if (commentIsNotAlone) {
                return;
            ***REMOVED***

            const previousTokenOrComment = sourceCode.getTokenBefore(token, { includeComments: true ***REMOVED***);
            const nextTokenOrComment = sourceCode.getTokenAfter(token, { includeComments: true ***REMOVED***);

            // check for newline before
            if (!exceptionStartAllowed && before && !commentAndEmptyLines.includes(prevLineNum) &&
                    !(astUtils.isCommentToken(previousTokenOrComment) && astUtils.isTokenOnSameLine(previousTokenOrComment, token))) {
                const lineStart = token.range[0] - token.loc.start.column;
                const range = [lineStart, lineStart];

                context.report({
                    node: token,
                    messageId: "before",
                    fix(fixer) {
                        return fixer.insertTextBeforeRange(range, "\n");
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***

            // check for newline after
            if (!exceptionEndAllowed && after && !commentAndEmptyLines.includes(nextLineNum) &&
                    !(astUtils.isCommentToken(nextTokenOrComment) && astUtils.isTokenOnSameLine(token, nextTokenOrComment))) {
                context.report({
                    node: token,
                    messageId: "after",
                    fix(fixer) {
                        return fixer.insertTextAfter(token, "\n");
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***

        ***REMOVED***

        //--------------------------------------------------------------------------
        // Public
        //--------------------------------------------------------------------------

        return {
            Program() {
                comments.forEach(token => {
                    if (token.type === "Line") {
                        if (options.beforeLineComment || options.afterLineComment) {
                            checkForEmptyLine(token, {
                                after: options.afterLineComment,
                                before: options.beforeLineComment
                            ***REMOVED***);
                        ***REMOVED***
                    ***REMOVED*** else if (token.type === "Block") {
                        if (options.beforeBlockComment || options.afterBlockComment) {
                            checkForEmptyLine(token, {
                                after: options.afterBlockComment,
                                before: options.beforeBlockComment
                            ***REMOVED***);
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
