// eslint/lib/rules/line-comment-position.js
/**
 * @fileoverview Rule to enforce the position of line comments
 * @author Alberto Rodríguez
 */
"use strict";

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "layout",

        docs: {
            description: "enforce position of line comments",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/line-comment-position"
        ***REMOVED***,

        schema: [
            {
                oneOf: [
                    {
                        enum: ["above", "beside"]
                    ***REMOVED***,
                    {
                        type: "object",
                        properties: {
                            position: {
                                enum: ["above", "beside"]
                            ***REMOVED***,
                            ignorePattern: {
                                type: "string"
                            ***REMOVED***,
                            applyDefaultPatterns: {
                                type: "boolean"
                            ***REMOVED***,
                            applyDefaultIgnorePatterns: {
                                type: "boolean"
                            ***REMOVED***
                        ***REMOVED***,
                        additionalProperties: false
                    ***REMOVED***
                ]
            ***REMOVED***
        ],
        messages: {
            above: "Expected comment to be above code.",
            beside: "Expected comment to be beside code."
        ***REMOVED***
    ***REMOVED***,

    create(context) {
        const options = context.options[0];

        let above,
            ignorePattern,
            applyDefaultIgnorePatterns = true;

        if (!options || typeof options === "string") {
            above = !options || options === "above";

        ***REMOVED*** else {
            above = !options.position || options.position === "above";
            ignorePattern = options.ignorePattern;

            if (Object.prototype.hasOwnProperty.call(options, "applyDefaultIgnorePatterns")) {
                applyDefaultIgnorePatterns = options.applyDefaultIgnorePatterns;
            ***REMOVED*** else {
                applyDefaultIgnorePatterns = options.applyDefaultPatterns !== false;
            ***REMOVED***
        ***REMOVED***

        const defaultIgnoreRegExp = astUtils.COMMENTS_IGNORE_PATTERN;
        const fallThroughRegExp = /^\s*falls?\s?through/u;
        const customIgnoreRegExp = new RegExp(ignorePattern, "u");
        const sourceCode = context.getSourceCode();

        //--------------------------------------------------------------------------
        // Public
        //--------------------------------------------------------------------------

        return {
            Program() {
                const comments = sourceCode.getAllComments();

                comments.filter(token => token.type === "Line").forEach(node => {
                    if (applyDefaultIgnorePatterns && (defaultIgnoreRegExp.test(node.value) || fallThroughRegExp.test(node.value))) {
                        return;
                    ***REMOVED***

                    if (ignorePattern && customIgnoreRegExp.test(node.value)) {
                        return;
                    ***REMOVED***

                    const previous = sourceCode.getTokenBefore(node, { includeComments: true ***REMOVED***);
                    const isOnSameLine = previous && previous.loc.end.line === node.loc.start.line;

                    if (above) {
                        if (isOnSameLine) {
                            context.report({
                                node,
                                messageId: "above"
                            ***REMOVED***);
                        ***REMOVED***
                    ***REMOVED*** else {
                        if (!isOnSameLine) {
                            context.report({
                                node,
                                messageId: "beside"
                            ***REMOVED***);
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
