// eslint/lib/rules/max-lines-per-function.js
/**
 * @fileoverview A rule to set the maximum number of line of code in a function.
 * @author Pete Ward <peteward44@gmail.com>
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");
const { upperCaseFirst ***REMOVED*** = require("../shared/string-utils");

//------------------------------------------------------------------------------
// Constants
//------------------------------------------------------------------------------

const OPTIONS_SCHEMA = {
    type: "object",
    properties: {
        max: {
            type: "integer",
            minimum: 0
        ***REMOVED***,
        skipComments: {
            type: "boolean"
        ***REMOVED***,
        skipBlankLines: {
            type: "boolean"
        ***REMOVED***,
        IIFEs: {
            type: "boolean"
        ***REMOVED***
    ***REMOVED***,
    additionalProperties: false
***REMOVED***;

const OPTIONS_OR_INTEGER_SCHEMA = {
    oneOf: [
        OPTIONS_SCHEMA,
        {
            type: "integer",
            minimum: 1
        ***REMOVED***
    ]
***REMOVED***;

/**
 * Given a list of comment nodes, return a map with numeric keys (source code line numbers) and comment token values.
 * @param {Array***REMOVED*** comments An array of comment nodes.
 * @returns {Map<string, Node>***REMOVED*** A map with numeric keys (source code line numbers) and comment token values.
 */
function getCommentLineNumbers(comments) {
    const map = new Map();

    comments.forEach(comment => {
        for (let i = comment.loc.start.line; i <= comment.loc.end.line; i++) {
            map.set(i, comment);
        ***REMOVED***
    ***REMOVED***);
    return map;
***REMOVED***

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

/** @type {import('../shared/types').Rule***REMOVED*** */
module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "Enforce a maximum number of lines of code in a function",
            recommended: false,
            url: "https://eslint.org/docs/latest/rules/max-lines-per-function"
        ***REMOVED***,

        schema: [
            OPTIONS_OR_INTEGER_SCHEMA
        ],
        messages: {
            exceed: "{{name***REMOVED******REMOVED*** has too many lines ({{lineCount***REMOVED******REMOVED***). Maximum allowed is {{maxLines***REMOVED******REMOVED***."
        ***REMOVED***
    ***REMOVED***,

    create(context) {
        const sourceCode = context.sourceCode;
        const lines = sourceCode.lines;

        const option = context.options[0];
        let maxLines = 50;
        let skipComments = false;
        let skipBlankLines = false;
        let IIFEs = false;

        if (typeof option === "object") {
            maxLines = typeof option.max === "number" ? option.max : 50;
            skipComments = !!option.skipComments;
            skipBlankLines = !!option.skipBlankLines;
            IIFEs = !!option.IIFEs;
        ***REMOVED*** else if (typeof option === "number") {
            maxLines = option;
        ***REMOVED***

        const commentLineNumbers = getCommentLineNumbers(sourceCode.getAllComments());

        //--------------------------------------------------------------------------
        // Helpers
        //--------------------------------------------------------------------------

        /**
         * Tells if a comment encompasses the entire line.
         * @param {string***REMOVED*** line The source line with a trailing comment
         * @param {number***REMOVED*** lineNumber The one-indexed line number this is on
         * @param {ASTNode***REMOVED*** comment The comment to remove
         * @returns {boolean***REMOVED*** If the comment covers the entire line
         */
        function isFullLineComment(line, lineNumber, comment) {
            const start = comment.loc.start,
                end = comment.loc.end,
                isFirstTokenOnLine = start.line === lineNumber && !line.slice(0, start.column).trim(),
                isLastTokenOnLine = end.line === lineNumber && !line.slice(end.column).trim();

            return comment &&
                (start.line < lineNumber || isFirstTokenOnLine) &&
                (end.line > lineNumber || isLastTokenOnLine);
        ***REMOVED***

        /**
         * Identifies is a node is a FunctionExpression which is part of an IIFE
         * @param {ASTNode***REMOVED*** node Node to test
         * @returns {boolean***REMOVED*** True if it's an IIFE
         */
        function isIIFE(node) {
            return (node.type === "FunctionExpression" || node.type === "ArrowFunctionExpression") && node.parent && node.parent.type === "CallExpression" && node.parent.callee === node;
        ***REMOVED***

        /**
         * Identifies is a node is a FunctionExpression which is embedded within a MethodDefinition or Property
         * @param {ASTNode***REMOVED*** node Node to test
         * @returns {boolean***REMOVED*** True if it's a FunctionExpression embedded within a MethodDefinition or Property
         */
        function isEmbedded(node) {
            if (!node.parent) {
                return false;
            ***REMOVED***
            if (node !== node.parent.value) {
                return false;
            ***REMOVED***
            if (node.parent.type === "MethodDefinition") {
                return true;
            ***REMOVED***
            if (node.parent.type === "Property") {
                return node.parent.method === true || node.parent.kind === "get" || node.parent.kind === "set";
            ***REMOVED***
            return false;
        ***REMOVED***

        /**
         * Count the lines in the function
         * @param {ASTNode***REMOVED*** funcNode Function AST node
         * @returns {void***REMOVED***
         * @private
         */
        function processFunction(funcNode) {
            const node = isEmbedded(funcNode) ? funcNode.parent : funcNode;

            if (!IIFEs && isIIFE(node)) {
                return;
            ***REMOVED***
            let lineCount = 0;

            for (let i = node.loc.start.line - 1; i < node.loc.end.line; ++i) {
                const line = lines[i];

                if (skipComments) {
                    if (commentLineNumbers.has(i + 1) && isFullLineComment(line, i + 1, commentLineNumbers.get(i + 1))) {
                        continue;
                    ***REMOVED***
                ***REMOVED***

                if (skipBlankLines) {
                    if (line.match(/^\s*$/u)) {
                        continue;
                    ***REMOVED***
                ***REMOVED***

                lineCount++;
            ***REMOVED***

            if (lineCount > maxLines) {
                const name = upperCaseFirst(astUtils.getFunctionNameWithKind(funcNode));

                context.report({
                    node,
                    messageId: "exceed",
                    data: { name, lineCount, maxLines ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        //--------------------------------------------------------------------------
        // Public API
        //--------------------------------------------------------------------------

        return {
            FunctionDeclaration: processFunction,
            FunctionExpression: processFunction,
            ArrowFunctionExpression: processFunction
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
