// eslint/lib/rules/newline-before-return.js
/**
 * @fileoverview Rule to require newlines before `return` statement
 * @author Kai Cataldo
 * @deprecated
 */
"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "layout",

        docs: {
            description: "require an empty line before `return` statements",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/newline-before-return"
        ***REMOVED***,

        fixable: "whitespace",
        schema: [],
        messages: {
            expected: "Expected newline before return statement."
        ***REMOVED***,

        deprecated: true,
        replacedBy: ["padding-line-between-statements"]
    ***REMOVED***,

    create(context) {
        const sourceCode = context.getSourceCode();

        //--------------------------------------------------------------------------
        // Helpers
        //--------------------------------------------------------------------------

        /**
         * Tests whether node is preceded by supplied tokens
         * @param {ASTNode***REMOVED*** node node to check
         * @param {Array***REMOVED*** testTokens array of tokens to test against
         * @returns {boolean***REMOVED*** Whether or not the node is preceded by one of the supplied tokens
         * @private
         */
        function isPrecededByTokens(node, testTokens) {
            const tokenBefore = sourceCode.getTokenBefore(node);

            return testTokens.some(token => tokenBefore.value === token);
        ***REMOVED***

        /**
         * Checks whether node is the first node after statement or in block
         * @param {ASTNode***REMOVED*** node node to check
         * @returns {boolean***REMOVED*** Whether or not the node is the first node after statement or in block
         * @private
         */
        function isFirstNode(node) {
            const parentType = node.parent.type;

            if (node.parent.body) {
                return Array.isArray(node.parent.body)
                    ? node.parent.body[0] === node
                    : node.parent.body === node;
            ***REMOVED***

            if (parentType === "IfStatement") {
                return isPrecededByTokens(node, ["else", ")"]);
            ***REMOVED***
            if (parentType === "DoWhileStatement") {
                return isPrecededByTokens(node, ["do"]);
            ***REMOVED***
            if (parentType === "SwitchCase") {
                return isPrecededByTokens(node, [":"]);
            ***REMOVED***
            return isPrecededByTokens(node, [")"]);

        ***REMOVED***

        /**
         * Returns the number of lines of comments that precede the node
         * @param {ASTNode***REMOVED*** node node to check for overlapping comments
         * @param {number***REMOVED*** lineNumTokenBefore line number of previous token, to check for overlapping comments
         * @returns {number***REMOVED*** Number of lines of comments that precede the node
         * @private
         */
        function calcCommentLines(node, lineNumTokenBefore) {
            const comments = sourceCode.getCommentsBefore(node);
            let numLinesComments = 0;

            if (!comments.length) {
                return numLinesComments;
            ***REMOVED***

            comments.forEach(comment => {
                numLinesComments++;

                if (comment.type === "Block") {
                    numLinesComments += comment.loc.end.line - comment.loc.start.line;
                ***REMOVED***

                // avoid counting lines with inline comments twice
                if (comment.loc.start.line === lineNumTokenBefore) {
                    numLinesComments--;
                ***REMOVED***

                if (comment.loc.end.line === node.loc.start.line) {
                    numLinesComments--;
                ***REMOVED***
            ***REMOVED***);

            return numLinesComments;
        ***REMOVED***

        /**
         * Returns the line number of the token before the node that is passed in as an argument
         * @param {ASTNode***REMOVED*** node The node to use as the start of the calculation
         * @returns {number***REMOVED*** Line number of the token before `node`
         * @private
         */
        function getLineNumberOfTokenBefore(node) {
            const tokenBefore = sourceCode.getTokenBefore(node);
            let lineNumTokenBefore;

            /**
             * Global return (at the beginning of a script) is a special case.
             * If there is no token before `return`, then we expect no line
             * break before the return. Comments are allowed to occupy lines
             * before the global return, just no blank lines.
             * Setting lineNumTokenBefore to zero in that case results in the
             * desired behavior.
             */
            if (tokenBefore) {
                lineNumTokenBefore = tokenBefore.loc.end.line;
            ***REMOVED*** else {
                lineNumTokenBefore = 0; // global return at beginning of script
            ***REMOVED***

            return lineNumTokenBefore;
        ***REMOVED***

        /**
         * Checks whether node is preceded by a newline
         * @param {ASTNode***REMOVED*** node node to check
         * @returns {boolean***REMOVED*** Whether or not the node is preceded by a newline
         * @private
         */
        function hasNewlineBefore(node) {
            const lineNumNode = node.loc.start.line;
            const lineNumTokenBefore = getLineNumberOfTokenBefore(node);
            const commentLines = calcCommentLines(node, lineNumTokenBefore);

            return (lineNumNode - lineNumTokenBefore - commentLines) > 1;
        ***REMOVED***

        /**
         * Checks whether it is safe to apply a fix to a given return statement.
         *
         * The fix is not considered safe if the given return statement has leading comments,
         * as we cannot safely determine if the newline should be added before or after the comments.
         * For more information, see: https://github.com/eslint/eslint/issues/5958#issuecomment-222767211
         * @param {ASTNode***REMOVED*** node The return statement node to check.
         * @returns {boolean***REMOVED*** `true` if it can fix the node.
         * @private
         */
        function canFix(node) {
            const leadingComments = sourceCode.getCommentsBefore(node);
            const lastLeadingComment = leadingComments[leadingComments.length - 1];
            const tokenBefore = sourceCode.getTokenBefore(node);

            if (leadingComments.length === 0) {
                return true;
            ***REMOVED***

            /*
             * if the last leading comment ends in the same line as the previous token and
             * does not share a line with the `return` node, we can consider it safe to fix.
             * Example:
             * function a() {
             *     var b; //comment
             *     return;
             * ***REMOVED***
             */
            if (lastLeadingComment.loc.end.line === tokenBefore.loc.end.line &&
                lastLeadingComment.loc.end.line !== node.loc.start.line) {
                return true;
            ***REMOVED***

            return false;
        ***REMOVED***

        //--------------------------------------------------------------------------
        // Public
        //--------------------------------------------------------------------------

        return {
            ReturnStatement(node) {
                if (!isFirstNode(node) && !hasNewlineBefore(node)) {
                    context.report({
                        node,
                        messageId: "expected",
                        fix(fixer) {
                            if (canFix(node)) {
                                const tokenBefore = sourceCode.getTokenBefore(node);
                                const newlines = node.loc.start.line === tokenBefore.loc.end.line ? "\n\n" : "\n";

                                return fixer.insertTextBefore(node, newlines);
                            ***REMOVED***
                            return null;
                        ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
