// eslint/lib/rules/no-await-in-loop.js
/**
 * @fileoverview Rule to disallow uses of await inside of loops.
 * @author Nat Mote (nmote)
 */
"use strict";

/**
 * Check whether it should stop traversing ancestors at the given node.
 * @param {ASTNode***REMOVED*** node A node to check.
 * @returns {boolean***REMOVED*** `true` if it should stop traversing.
 */
function isBoundary(node) {
    const t = node.type;

    return (
        t === "FunctionDeclaration" ||
        t === "FunctionExpression" ||
        t === "ArrowFunctionExpression" ||

        /*
         * Don't report the await expressions on for-await-of loop since it's
         * asynchronous iteration intentionally.
         */
        (t === "ForOfStatement" && node.await === true)
    );
***REMOVED***

/**
 * Check whether the given node is in loop.
 * @param {ASTNode***REMOVED*** node A node to check.
 * @param {ASTNode***REMOVED*** parent A parent node to check.
 * @returns {boolean***REMOVED*** `true` if the node is in loop.
 */
function isLooped(node, parent) {
    switch (parent.type) {
        case "ForStatement":
            return (
                node === parent.test ||
                node === parent.update ||
                node === parent.body
            );

        case "ForOfStatement":
        case "ForInStatement":
            return node === parent.body;

        case "WhileStatement":
        case "DoWhileStatement":
            return node === parent.test || node === parent.body;

        default:
            return false;
    ***REMOVED***
***REMOVED***

module.exports = {
    meta: {
        type: "problem",

        docs: {
            description: "disallow `await` inside of loops",
            category: "Possible Errors",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-await-in-loop"
        ***REMOVED***,

        schema: [],

        messages: {
            unexpectedAwait: "Unexpected `await` inside a loop."
        ***REMOVED***
    ***REMOVED***,
    create(context) {

        /**
         * Validate an await expression.
         * @param {ASTNode***REMOVED*** awaitNode An AwaitExpression or ForOfStatement node to validate.
         * @returns {void***REMOVED***
         */
        function validate(awaitNode) {
            if (awaitNode.type === "ForOfStatement" && !awaitNode.await) {
                return;
            ***REMOVED***

            let node = awaitNode;
            let parent = node.parent;

            while (parent && !isBoundary(parent)) {
                if (isLooped(node, parent)) {
                    context.report({
                        node: awaitNode,
                        messageId: "unexpectedAwait"
                    ***REMOVED***);
                    return;
                ***REMOVED***
                node = parent;
                parent = parent.parent;
            ***REMOVED***
        ***REMOVED***

        return {
            AwaitExpression: validate,
            ForOfStatement: validate
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
