// eslint/lib/rules/no-array-constructor.js
/**
 * @fileoverview Disallow construction of dense arrays using the Array constructor
 * @author Matt DuVall <http://www.mattduvall.com/>
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "disallow `Array` constructors",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-array-constructor"
        ***REMOVED***,

        schema: [],

        messages: {
            preferLiteral: "The array literal notation [] is preferable."
        ***REMOVED***
    ***REMOVED***,

    create(context) {

        /**
         * Disallow construction of dense arrays using the Array constructor
         * @param {ASTNode***REMOVED*** node node to evaluate
         * @returns {void***REMOVED***
         * @private
         */
        function check(node) {
            if (
                node.arguments.length !== 1 &&
                node.callee.type === "Identifier" &&
                node.callee.name === "Array"
            ) {
                context.report({ node, messageId: "preferLiteral" ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        return {
            CallExpression: check,
            NewExpression: check
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
