// eslint/lib/rules/no-bitwise.js
/**
 * @fileoverview Rule to flag bitwise identifiers
 * @author Nicholas C. Zakas
 */

"use strict";

/*
 *
 * Set of bitwise operators.
 *
 */
const BITWISE_OPERATORS = [
    "^", "|", "&", "<<", ">>", ">>>",
    "^=", "|=", "&=", "<<=", ">>=", ">>>=",
    "~"
];

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "disallow bitwise operators",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-bitwise"
        ***REMOVED***,

        schema: [
            {
                type: "object",
                properties: {
                    allow: {
                        type: "array",
                        items: {
                            enum: BITWISE_OPERATORS
                        ***REMOVED***,
                        uniqueItems: true
                    ***REMOVED***,
                    int32Hint: {
                        type: "boolean",
                        default: false
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],

        messages: {
            unexpected: "Unexpected use of '{{operator***REMOVED******REMOVED***'."
        ***REMOVED***
    ***REMOVED***,

    create(context) {
        const options = context.options[0] || {***REMOVED***;
        const allowed = options.allow || [];
        const int32Hint = options.int32Hint === true;

        /**
         * Reports an unexpected use of a bitwise operator.
         * @param   {ASTNode***REMOVED*** node Node which contains the bitwise operator.
         * @returns {void***REMOVED***
         */
        function report(node) {
            context.report({ node, messageId: "unexpected", data: { operator: node.operator ***REMOVED*** ***REMOVED***);
        ***REMOVED***

        /**
         * Checks if the given node has a bitwise operator.
         * @param   {ASTNode***REMOVED*** node The node to check.
         * @returns {boolean***REMOVED*** Whether or not the node has a bitwise operator.
         */
        function hasBitwiseOperator(node) {
            return BITWISE_OPERATORS.indexOf(node.operator) !== -1;
        ***REMOVED***

        /**
         * Checks if exceptions were provided, e.g. `{ allow: ['~', '|'] ***REMOVED***`.
         * @param   {ASTNode***REMOVED*** node The node to check.
         * @returns {boolean***REMOVED*** Whether or not the node has a bitwise operator.
         */
        function allowedOperator(node) {
            return allowed.indexOf(node.operator) !== -1;
        ***REMOVED***

        /**
         * Checks if the given bitwise operator is used for integer typecasting, i.e. "|0"
         * @param   {ASTNode***REMOVED*** node The node to check.
         * @returns {boolean***REMOVED*** whether the node is used in integer typecasting.
         */
        function isInt32Hint(node) {
            return int32Hint && node.operator === "|" && node.right &&
              node.right.type === "Literal" && node.right.value === 0;
        ***REMOVED***

        /**
         * Report if the given node contains a bitwise operator.
         * @param   {ASTNode***REMOVED*** node The node to check.
         * @returns {void***REMOVED***
         */
        function checkNodeForBitwiseOperator(node) {
            if (hasBitwiseOperator(node) && !allowedOperator(node) && !isInt32Hint(node)) {
                report(node);
            ***REMOVED***
        ***REMOVED***

        return {
            AssignmentExpression: checkNodeForBitwiseOperator,
            BinaryExpression: checkNodeForBitwiseOperator,
            UnaryExpression: checkNodeForBitwiseOperator
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
