// eslint/lib/rules/no-ex-assign.js
/**
 * @fileoverview Rule to flag assignment of the exception parameter
 * @author Stephen Murray <spmurrayzzz>
 */

"use strict";

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

/** @type {import('../shared/types').Rule***REMOVED*** */
module.exports = {
    meta: {
        type: "problem",

        docs: {
            description: "Disallow reassigning exceptions in `catch` clauses",
            recommended: true,
            url: "https://eslint.org/docs/latest/rules/no-ex-assign"
        ***REMOVED***,

        schema: [],

        messages: {
            unexpected: "Do not assign to the exception parameter."
        ***REMOVED***
    ***REMOVED***,

    create(context) {

        const sourceCode = context.sourceCode;

        /**
         * Finds and reports references that are non initializer and writable.
         * @param {Variable***REMOVED*** variable A variable to check.
         * @returns {void***REMOVED***
         */
        function checkVariable(variable) {
            astUtils.getModifyingReferences(variable.references).forEach(reference => {
                context.report({ node: reference.identifier, messageId: "unexpected" ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***

        return {
            CatchClause(node) {
                sourceCode.getDeclaredVariables(node).forEach(checkVariable);
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
