// eslint/lib/rules/no-extra-label.js
/**
 * @fileoverview Rule to disallow unnecessary labels
 * @author Toru Nagashima
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "disallow unnecessary labels",
            category: "Best Practices",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-extra-label"
        ***REMOVED***,

        schema: [],
        fixable: "code",

        messages: {
            unexpected: "This label '{{name***REMOVED******REMOVED***' is unnecessary."
        ***REMOVED***
    ***REMOVED***,

    create(context) {
        const sourceCode = context.getSourceCode();
        let scopeInfo = null;

        /**
         * Creates a new scope with a breakable statement.
         * @param {ASTNode***REMOVED*** node A node to create. This is a BreakableStatement.
         * @returns {void***REMOVED***
         */
        function enterBreakableStatement(node) {
            scopeInfo = {
                label: node.parent.type === "LabeledStatement" ? node.parent.label : null,
                breakable: true,
                upper: scopeInfo
            ***REMOVED***;
        ***REMOVED***

        /**
         * Removes the top scope of the stack.
         * @returns {void***REMOVED***
         */
        function exitBreakableStatement() {
            scopeInfo = scopeInfo.upper;
        ***REMOVED***

        /**
         * Creates a new scope with a labeled statement.
         *
         * This ignores it if the body is a breakable statement.
         * In this case it's handled in the `enterBreakableStatement` function.
         * @param {ASTNode***REMOVED*** node A node to create. This is a LabeledStatement.
         * @returns {void***REMOVED***
         */
        function enterLabeledStatement(node) {
            if (!astUtils.isBreakableStatement(node.body)) {
                scopeInfo = {
                    label: node.label,
                    breakable: false,
                    upper: scopeInfo
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***

        /**
         * Removes the top scope of the stack.
         *
         * This ignores it if the body is a breakable statement.
         * In this case it's handled in the `exitBreakableStatement` function.
         * @param {ASTNode***REMOVED*** node A node. This is a LabeledStatement.
         * @returns {void***REMOVED***
         */
        function exitLabeledStatement(node) {
            if (!astUtils.isBreakableStatement(node.body)) {
                scopeInfo = scopeInfo.upper;
            ***REMOVED***
        ***REMOVED***

        /**
         * Reports a given control node if it's unnecessary.
         * @param {ASTNode***REMOVED*** node A node. This is a BreakStatement or a
         *      ContinueStatement.
         * @returns {void***REMOVED***
         */
        function reportIfUnnecessary(node) {
            if (!node.label) {
                return;
            ***REMOVED***

            const labelNode = node.label;

            for (let info = scopeInfo; info !== null; info = info.upper) {
                if (info.breakable || info.label && info.label.name === labelNode.name) {
                    if (info.breakable && info.label && info.label.name === labelNode.name) {
                        context.report({
                            node: labelNode,
                            messageId: "unexpected",
                            data: labelNode,
                            fix(fixer) {
                                const breakOrContinueToken = sourceCode.getFirstToken(node);

                                if (sourceCode.commentsExistBetween(breakOrContinueToken, labelNode)) {
                                    return null;
                                ***REMOVED***

                                return fixer.removeRange([breakOrContinueToken.range[1], labelNode.range[1]]);
                            ***REMOVED***
                        ***REMOVED***);
                    ***REMOVED***
                    return;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        return {
            WhileStatement: enterBreakableStatement,
            "WhileStatement:exit": exitBreakableStatement,
            DoWhileStatement: enterBreakableStatement,
            "DoWhileStatement:exit": exitBreakableStatement,
            ForStatement: enterBreakableStatement,
            "ForStatement:exit": exitBreakableStatement,
            ForInStatement: enterBreakableStatement,
            "ForInStatement:exit": exitBreakableStatement,
            ForOfStatement: enterBreakableStatement,
            "ForOfStatement:exit": exitBreakableStatement,
            SwitchStatement: enterBreakableStatement,
            "SwitchStatement:exit": exitBreakableStatement,
            LabeledStatement: enterLabeledStatement,
            "LabeledStatement:exit": exitLabeledStatement,
            BreakStatement: reportIfUnnecessary,
            ContinueStatement: reportIfUnnecessary
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
