// eslint/lib/rules/no-extra-parens.js
/**
 * @fileoverview Disallow parenthesising higher precedence subexpressions.
 * @author Michael Ficarra
 */
"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

const { isParenthesized: isParenthesizedRaw ***REMOVED*** = require("eslint-utils");
const astUtils = require("./utils/ast-utils.js");

module.exports = {
    meta: {
        type: "layout",

        docs: {
            description: "disallow unnecessary parentheses",
            category: "Possible Errors",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-extra-parens"
        ***REMOVED***,

        fixable: "code",

        schema: {
            anyOf: [
                {
                    type: "array",
                    items: [
                        {
                            enum: ["functions"]
                        ***REMOVED***
                    ],
                    minItems: 0,
                    maxItems: 1
                ***REMOVED***,
                {
                    type: "array",
                    items: [
                        {
                            enum: ["all"]
                        ***REMOVED***,
                        {
                            type: "object",
                            properties: {
                                conditionalAssign: { type: "boolean" ***REMOVED***,
                                nestedBinaryExpressions: { type: "boolean" ***REMOVED***,
                                returnAssign: { type: "boolean" ***REMOVED***,
                                ignoreJSX: { enum: ["none", "all", "single-line", "multi-line"] ***REMOVED***,
                                enforceForArrowConditionals: { type: "boolean" ***REMOVED***,
                                enforceForSequenceExpressions: { type: "boolean" ***REMOVED***,
                                enforceForNewInMemberExpressions: { type: "boolean" ***REMOVED***,
                                enforceForFunctionPrototypeMethods: { type: "boolean" ***REMOVED***
                            ***REMOVED***,
                            additionalProperties: false
                        ***REMOVED***
                    ],
                    minItems: 0,
                    maxItems: 2
                ***REMOVED***
            ]
        ***REMOVED***,

        messages: {
            unexpected: "Unnecessary parentheses around expression."
        ***REMOVED***
    ***REMOVED***,

    create(context) {
        const sourceCode = context.getSourceCode();

        const tokensToIgnore = new WeakSet();
        const precedence = astUtils.getPrecedence;
        const ALL_NODES = context.options[0] !== "functions";
        const EXCEPT_COND_ASSIGN = ALL_NODES && context.options[1] && context.options[1].conditionalAssign === false;
        const NESTED_BINARY = ALL_NODES && context.options[1] && context.options[1].nestedBinaryExpressions === false;
        const EXCEPT_RETURN_ASSIGN = ALL_NODES && context.options[1] && context.options[1].returnAssign === false;
        const IGNORE_JSX = ALL_NODES && context.options[1] && context.options[1].ignoreJSX;
        const IGNORE_ARROW_CONDITIONALS = ALL_NODES && context.options[1] &&
            context.options[1].enforceForArrowConditionals === false;
        const IGNORE_SEQUENCE_EXPRESSIONS = ALL_NODES && context.options[1] &&
            context.options[1].enforceForSequenceExpressions === false;
        const IGNORE_NEW_IN_MEMBER_EXPR = ALL_NODES && context.options[1] &&
            context.options[1].enforceForNewInMemberExpressions === false;
        const IGNORE_FUNCTION_PROTOTYPE_METHODS = ALL_NODES && context.options[1] &&
            context.options[1].enforceForFunctionPrototypeMethods === false;

        const PRECEDENCE_OF_ASSIGNMENT_EXPR = precedence({ type: "AssignmentExpression" ***REMOVED***);
        const PRECEDENCE_OF_UPDATE_EXPR = precedence({ type: "UpdateExpression" ***REMOVED***);

        let reportsBuffer;

        /**
         * Determines whether the given node is a `call` or `apply` method call, invoked directly on a `FunctionExpression` node.
         * Example: function(){***REMOVED***.call()
         * @param {ASTNode***REMOVED*** node The node to be checked.
         * @returns {boolean***REMOVED*** True if the node is an immediate `call` or `apply` method call.
         * @private
         */
        function isImmediateFunctionPrototypeMethodCall(node) {
            const callNode = astUtils.skipChainExpression(node);

            if (callNode.type !== "CallExpression") {
                return false;
            ***REMOVED***
            const callee = astUtils.skipChainExpression(callNode.callee);

            return (
                callee.type === "MemberExpression" &&
                callee.object.type === "FunctionExpression" &&
                ["call", "apply"].includes(astUtils.getStaticPropertyName(callee))
            );
        ***REMOVED***

        /**
         * Determines if this rule should be enforced for a node given the current configuration.
         * @param {ASTNode***REMOVED*** node The node to be checked.
         * @returns {boolean***REMOVED*** True if the rule should be enforced for this node.
         * @private
         */
        function ruleApplies(node) {
            if (node.type === "JSXElement" || node.type === "JSXFragment") {
                const isSingleLine = node.loc.start.line === node.loc.end.line;

                switch (IGNORE_JSX) {

                    // Exclude this JSX element from linting
                    case "all":
                        return false;

                    // Exclude this JSX element if it is multi-line element
                    case "multi-line":
                        return isSingleLine;

                    // Exclude this JSX element if it is single-line element
                    case "single-line":
                        return !isSingleLine;

                    // Nothing special to be done for JSX elements
                    case "none":
                        break;

                    // no default
                ***REMOVED***
            ***REMOVED***

            if (node.type === "SequenceExpression" && IGNORE_SEQUENCE_EXPRESSIONS) {
                return false;
            ***REMOVED***

            if (isImmediateFunctionPrototypeMethodCall(node) && IGNORE_FUNCTION_PROTOTYPE_METHODS) {
                return false;
            ***REMOVED***

            return ALL_NODES || node.type === "FunctionExpression" || node.type === "ArrowFunctionExpression";
        ***REMOVED***

        /**
         * Determines if a node is surrounded by parentheses.
         * @param {ASTNode***REMOVED*** node The node to be checked.
         * @returns {boolean***REMOVED*** True if the node is parenthesised.
         * @private
         */
        function isParenthesised(node) {
            return isParenthesizedRaw(1, node, sourceCode);
        ***REMOVED***

        /**
         * Determines if a node is surrounded by parentheses twice.
         * @param {ASTNode***REMOVED*** node The node to be checked.
         * @returns {boolean***REMOVED*** True if the node is doubly parenthesised.
         * @private
         */
        function isParenthesisedTwice(node) {
            return isParenthesizedRaw(2, node, sourceCode);
        ***REMOVED***

        /**
         * Determines if a node is surrounded by (potentially) invalid parentheses.
         * @param {ASTNode***REMOVED*** node The node to be checked.
         * @returns {boolean***REMOVED*** True if the node is incorrectly parenthesised.
         * @private
         */
        function hasExcessParens(node) {
            return ruleApplies(node) && isParenthesised(node);
        ***REMOVED***

        /**
         * Determines if a node that is expected to be parenthesised is surrounded by
         * (potentially) invalid extra parentheses.
         * @param {ASTNode***REMOVED*** node The node to be checked.
         * @returns {boolean***REMOVED*** True if the node is has an unexpected extra pair of parentheses.
         * @private
         */
        function hasDoubleExcessParens(node) {
            return ruleApplies(node) && isParenthesisedTwice(node);
        ***REMOVED***

        /**
         * Determines if a node that is expected to be parenthesised is surrounded by
         * (potentially) invalid extra parentheses with considering precedence level of the node.
         * If the preference level of the node is not higher or equal to precedence lower limit, it also checks
         * whether the node is surrounded by parentheses twice or not.
         * @param {ASTNode***REMOVED*** node The node to be checked.
         * @param {number***REMOVED*** precedenceLowerLimit The lower limit of precedence.
         * @returns {boolean***REMOVED*** True if the node is has an unexpected extra pair of parentheses.
         * @private
         */
        function hasExcessParensWithPrecedence(node, precedenceLowerLimit) {
            if (ruleApplies(node) && isParenthesised(node)) {
                if (
                    precedence(node) >= precedenceLowerLimit ||
                    isParenthesisedTwice(node)
                ) {
                    return true;
                ***REMOVED***
            ***REMOVED***
            return false;
        ***REMOVED***

        /**
         * Determines if a node test expression is allowed to have a parenthesised assignment
         * @param {ASTNode***REMOVED*** node The node to be checked.
         * @returns {boolean***REMOVED*** True if the assignment can be parenthesised.
         * @private
         */
        function isCondAssignException(node) {
            return EXCEPT_COND_ASSIGN && node.test.type === "AssignmentExpression";
        ***REMOVED***

        /**
         * Determines if a node is in a return statement
         * @param {ASTNode***REMOVED*** node The node to be checked.
         * @returns {boolean***REMOVED*** True if the node is in a return statement.
         * @private
         */
        function isInReturnStatement(node) {
            for (let currentNode = node; currentNode; currentNode = currentNode.parent) {
                if (
                    currentNode.type === "ReturnStatement" ||
                    (currentNode.type === "ArrowFunctionExpression" && currentNode.body.type !== "BlockStatement")
                ) {
                    return true;
                ***REMOVED***
            ***REMOVED***

            return false;
        ***REMOVED***

        /**
         * Determines if a constructor function is newed-up with parens
         * @param {ASTNode***REMOVED*** newExpression The NewExpression node to be checked.
         * @returns {boolean***REMOVED*** True if the constructor is called with parens.
         * @private
         */
        function isNewExpressionWithParens(newExpression) {
            const lastToken = sourceCode.getLastToken(newExpression);
            const penultimateToken = sourceCode.getTokenBefore(lastToken);

            return newExpression.arguments.length > 0 ||
                (

                    // The expression should end with its own parens, e.g., new new foo() is not a new expression with parens
                    astUtils.isOpeningParenToken(penultimateToken) &&
                    astUtils.isClosingParenToken(lastToken) &&
                    newExpression.callee.range[1] < newExpression.range[1]
                );
        ***REMOVED***

        /**
         * Determines if a node is or contains an assignment expression
         * @param {ASTNode***REMOVED*** node The node to be checked.
         * @returns {boolean***REMOVED*** True if the node is or contains an assignment expression.
         * @private
         */
        function containsAssignment(node) {
            if (node.type === "AssignmentExpression") {
                return true;
            ***REMOVED***
            if (node.type === "ConditionalExpression" &&
                    (node.consequent.type === "AssignmentExpression" || node.alternate.type === "AssignmentExpression")) {
                return true;
            ***REMOVED***
            if ((node.left && node.left.type === "AssignmentExpression") ||
                    (node.right && node.right.type === "AssignmentExpression")) {
                return true;
            ***REMOVED***

            return false;
        ***REMOVED***

        /**
         * Determines if a node is contained by or is itself a return statement and is allowed to have a parenthesised assignment
         * @param {ASTNode***REMOVED*** node The node to be checked.
         * @returns {boolean***REMOVED*** True if the assignment can be parenthesised.
         * @private
         */
        function isReturnAssignException(node) {
            if (!EXCEPT_RETURN_ASSIGN || !isInReturnStatement(node)) {
                return false;
            ***REMOVED***

            if (node.type === "ReturnStatement") {
                return node.argument && containsAssignment(node.argument);
            ***REMOVED***
            if (node.type === "ArrowFunctionExpression" && node.body.type !== "BlockStatement") {
                return containsAssignment(node.body);
            ***REMOVED***
            return containsAssignment(node);

        ***REMOVED***

        /**
         * Determines if a node following a [no LineTerminator here] restriction is
         * surrounded by (potentially) invalid extra parentheses.
         * @param {Token***REMOVED*** token The token preceding the [no LineTerminator here] restriction.
         * @param {ASTNode***REMOVED*** node The node to be checked.
         * @returns {boolean***REMOVED*** True if the node is incorrectly parenthesised.
         * @private
         */
        function hasExcessParensNoLineTerminator(token, node) {
            if (token.loc.end.line === node.loc.start.line) {
                return hasExcessParens(node);
            ***REMOVED***

            return hasDoubleExcessParens(node);
        ***REMOVED***

        /**
         * Determines whether a node should be preceded by an additional space when removing parens
         * @param {ASTNode***REMOVED*** node node to evaluate; must be surrounded by parentheses
         * @returns {boolean***REMOVED*** `true` if a space should be inserted before the node
         * @private
         */
        function requiresLeadingSpace(node) {
            const leftParenToken = sourceCode.getTokenBefore(node);
            const tokenBeforeLeftParen = sourceCode.getTokenBefore(leftParenToken, { includeComments: true ***REMOVED***);
            const tokenAfterLeftParen = sourceCode.getTokenAfter(leftParenToken, { includeComments: true ***REMOVED***);

            return tokenBeforeLeftParen &&
                tokenBeforeLeftParen.range[1] === leftParenToken.range[0] &&
                leftParenToken.range[1] === tokenAfterLeftParen.range[0] &&
                !astUtils.canTokensBeAdjacent(tokenBeforeLeftParen, tokenAfterLeftParen);
        ***REMOVED***

        /**
         * Determines whether a node should be followed by an additional space when removing parens
         * @param {ASTNode***REMOVED*** node node to evaluate; must be surrounded by parentheses
         * @returns {boolean***REMOVED*** `true` if a space should be inserted after the node
         * @private
         */
        function requiresTrailingSpace(node) {
            const nextTwoTokens = sourceCode.getTokensAfter(node, { count: 2 ***REMOVED***);
            const rightParenToken = nextTwoTokens[0];
            const tokenAfterRightParen = nextTwoTokens[1];
            const tokenBeforeRightParen = sourceCode.getLastToken(node);

            return rightParenToken && tokenAfterRightParen &&
                !sourceCode.isSpaceBetweenTokens(rightParenToken, tokenAfterRightParen) &&
                !astUtils.canTokensBeAdjacent(tokenBeforeRightParen, tokenAfterRightParen);
        ***REMOVED***

        /**
         * Determines if a given expression node is an IIFE
         * @param {ASTNode***REMOVED*** node The node to check
         * @returns {boolean***REMOVED*** `true` if the given node is an IIFE
         */
        function isIIFE(node) {
            const maybeCallNode = astUtils.skipChainExpression(node);

            return maybeCallNode.type === "CallExpression" && maybeCallNode.callee.type === "FunctionExpression";
        ***REMOVED***

        /**
         * Determines if the given node can be the assignment target in destructuring or the LHS of an assignment.
         * This is to avoid an autofix that could change behavior because parsers mistakenly allow invalid syntax,
         * such as `(a = b) = c` and `[(a = b) = c] = []`. Ideally, this function shouldn't be necessary.
         * @param {ASTNode***REMOVED*** [node] The node to check
         * @returns {boolean***REMOVED*** `true` if the given node can be a valid assignment target
         */
        function canBeAssignmentTarget(node) {
            return node && (node.type === "Identifier" || node.type === "MemberExpression");
        ***REMOVED***

        /**
         * Report the node
         * @param {ASTNode***REMOVED*** node node to evaluate
         * @returns {void***REMOVED***
         * @private
         */
        function report(node) {
            const leftParenToken = sourceCode.getTokenBefore(node);
            const rightParenToken = sourceCode.getTokenAfter(node);

            if (!isParenthesisedTwice(node)) {
                if (tokensToIgnore.has(sourceCode.getFirstToken(node))) {
                    return;
                ***REMOVED***

                if (isIIFE(node) && !isParenthesised(node.callee)) {
                    return;
                ***REMOVED***
            ***REMOVED***

            /**
             * Finishes reporting
             * @returns {void***REMOVED***
             * @private
             */
            function finishReport() {
                context.report({
                    node,
                    loc: leftParenToken.loc,
                    messageId: "unexpected",
                    fix(fixer) {
                        const parenthesizedSource = sourceCode.text.slice(leftParenToken.range[1], rightParenToken.range[0]);

                        return fixer.replaceTextRange([
                            leftParenToken.range[0],
                            rightParenToken.range[1]
                        ], (requiresLeadingSpace(node) ? " " : "") + parenthesizedSource + (requiresTrailingSpace(node) ? " " : ""));
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***

            if (reportsBuffer) {
                reportsBuffer.reports.push({ node, finishReport ***REMOVED***);
                return;
            ***REMOVED***

            finishReport();
        ***REMOVED***

        /**
         * Evaluate a argument of the node.
         * @param {ASTNode***REMOVED*** node node to evaluate
         * @returns {void***REMOVED***
         * @private
         */
        function checkArgumentWithPrecedence(node) {
            if (hasExcessParensWithPrecedence(node.argument, precedence(node))) {
                report(node.argument);
            ***REMOVED***
        ***REMOVED***

        /**
         * Check if a member expression contains a call expression
         * @param {ASTNode***REMOVED*** node MemberExpression node to evaluate
         * @returns {boolean***REMOVED*** true if found, false if not
         */
        function doesMemberExpressionContainCallExpression(node) {
            let currentNode = node.object;
            let currentNodeType = node.object.type;

            while (currentNodeType === "MemberExpression") {
                currentNode = currentNode.object;
                currentNodeType = currentNode.type;
            ***REMOVED***

            return currentNodeType === "CallExpression";
        ***REMOVED***

        /**
         * Evaluate a new call
         * @param {ASTNode***REMOVED*** node node to evaluate
         * @returns {void***REMOVED***
         * @private
         */
        function checkCallNew(node) {
            const callee = node.callee;

            if (hasExcessParensWithPrecedence(callee, precedence(node))) {
                if (
                    hasDoubleExcessParens(callee) ||
                    !(
                        isIIFE(node) ||

                        // (new A)(); new (new A)();
                        (
                            callee.type === "NewExpression" &&
                            !isNewExpressionWithParens(callee) &&
                            !(
                                node.type === "NewExpression" &&
                                !isNewExpressionWithParens(node)
                            )
                        ) ||

                        // new (a().b)(); new (a.b().c);
                        (
                            node.type === "NewExpression" &&
                            callee.type === "MemberExpression" &&
                            doesMemberExpressionContainCallExpression(callee)
                        ) ||

                        // (a?.b)(); (a?.())();
                        (
                            !node.optional &&
                            callee.type === "ChainExpression"
                        )
                    )
                ) {
                    report(node.callee);
                ***REMOVED***
            ***REMOVED***
            node.arguments
                .filter(arg => hasExcessParensWithPrecedence(arg, PRECEDENCE_OF_ASSIGNMENT_EXPR))
                .forEach(report);
        ***REMOVED***

        /**
         * Evaluate binary logicals
         * @param {ASTNode***REMOVED*** node node to evaluate
         * @returns {void***REMOVED***
         * @private
         */
        function checkBinaryLogical(node) {
            const prec = precedence(node);
            const leftPrecedence = precedence(node.left);
            const rightPrecedence = precedence(node.right);
            const isExponentiation = node.operator === "**";
            const shouldSkipLeft = NESTED_BINARY && (node.left.type === "BinaryExpression" || node.left.type === "LogicalExpression");
            const shouldSkipRight = NESTED_BINARY && (node.right.type === "BinaryExpression" || node.right.type === "LogicalExpression");

            if (!shouldSkipLeft && hasExcessParens(node.left)) {
                if (
                    !(["AwaitExpression", "UnaryExpression"].includes(node.left.type) && isExponentiation) &&
                    !astUtils.isMixedLogicalAndCoalesceExpressions(node.left, node) &&
                    (leftPrecedence > prec || (leftPrecedence === prec && !isExponentiation)) ||
                    isParenthesisedTwice(node.left)
                ) {
                    report(node.left);
                ***REMOVED***
            ***REMOVED***

            if (!shouldSkipRight && hasExcessParens(node.right)) {
                if (
                    !astUtils.isMixedLogicalAndCoalesceExpressions(node.right, node) &&
                    (rightPrecedence > prec || (rightPrecedence === prec && isExponentiation)) ||
                    isParenthesisedTwice(node.right)
                ) {
                    report(node.right);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        /**
         * Check the parentheses around the super class of the given class definition.
         * @param {ASTNode***REMOVED*** node The node of class declarations to check.
         * @returns {void***REMOVED***
         */
        function checkClass(node) {
            if (!node.superClass) {
                return;
            ***REMOVED***

            /*
             * If `node.superClass` is a LeftHandSideExpression, parentheses are extra.
             * Otherwise, parentheses are needed.
             */
            const hasExtraParens = precedence(node.superClass) > PRECEDENCE_OF_UPDATE_EXPR
                ? hasExcessParens(node.superClass)
                : hasDoubleExcessParens(node.superClass);

            if (hasExtraParens) {
                report(node.superClass);
            ***REMOVED***
        ***REMOVED***

        /**
         * Check the parentheses around the argument of the given spread operator.
         * @param {ASTNode***REMOVED*** node The node of spread elements/properties to check.
         * @returns {void***REMOVED***
         */
        function checkSpreadOperator(node) {
            if (hasExcessParensWithPrecedence(node.argument, PRECEDENCE_OF_ASSIGNMENT_EXPR)) {
                report(node.argument);
            ***REMOVED***
        ***REMOVED***

        /**
         * Checks the parentheses for an ExpressionStatement or ExportDefaultDeclaration
         * @param {ASTNode***REMOVED*** node The ExpressionStatement.expression or ExportDefaultDeclaration.declaration node
         * @returns {void***REMOVED***
         */
        function checkExpressionOrExportStatement(node) {
            const firstToken = isParenthesised(node) ? sourceCode.getTokenBefore(node) : sourceCode.getFirstToken(node);
            const secondToken = sourceCode.getTokenAfter(firstToken, astUtils.isNotOpeningParenToken);
            const thirdToken = secondToken ? sourceCode.getTokenAfter(secondToken) : null;
            const tokenAfterClosingParens = secondToken ? sourceCode.getTokenAfter(secondToken, astUtils.isNotClosingParenToken) : null;

            if (
                astUtils.isOpeningParenToken(firstToken) &&
                (
                    astUtils.isOpeningBraceToken(secondToken) ||
                    secondToken.type === "Keyword" && (
                        secondToken.value === "function" ||
                        secondToken.value === "class" ||
                        secondToken.value === "let" &&
                            tokenAfterClosingParens &&
                            (
                                astUtils.isOpeningBracketToken(tokenAfterClosingParens) ||
                                tokenAfterClosingParens.type === "Identifier"
                            )
                    ) ||
                    secondToken && secondToken.type === "Identifier" && secondToken.value === "async" && thirdToken && thirdToken.type === "Keyword" && thirdToken.value === "function"
                )
            ) {
                tokensToIgnore.add(secondToken);
            ***REMOVED***

            const hasExtraParens = node.parent.type === "ExportDefaultDeclaration"
                ? hasExcessParensWithPrecedence(node, PRECEDENCE_OF_ASSIGNMENT_EXPR)
                : hasExcessParens(node);

            if (hasExtraParens) {
                report(node);
            ***REMOVED***
        ***REMOVED***

        /**
         * Finds the path from the given node to the specified ancestor.
         * @param {ASTNode***REMOVED*** node First node in the path.
         * @param {ASTNode***REMOVED*** ancestor Last node in the path.
         * @returns {ASTNode[]***REMOVED*** Path, including both nodes.
         * @throws {Error***REMOVED*** If the given node does not have the specified ancestor.
         */
        function pathToAncestor(node, ancestor) {
            const path = [node];
            let currentNode = node;

            while (currentNode !== ancestor) {

                currentNode = currentNode.parent;

                /* istanbul ignore if */
                if (currentNode === null) {
                    throw new Error("Nodes are not in the ancestor-descendant relationship.");
                ***REMOVED***

                path.push(currentNode);
            ***REMOVED***

            return path;
        ***REMOVED***

        /**
         * Finds the path from the given node to the specified descendant.
         * @param {ASTNode***REMOVED*** node First node in the path.
         * @param {ASTNode***REMOVED*** descendant Last node in the path.
         * @returns {ASTNode[]***REMOVED*** Path, including both nodes.
         * @throws {Error***REMOVED*** If the given node does not have the specified descendant.
         */
        function pathToDescendant(node, descendant) {
            return pathToAncestor(descendant, node).reverse();
        ***REMOVED***

        /**
         * Checks whether the syntax of the given ancestor of an 'in' expression inside a for-loop initializer
         * is preventing the 'in' keyword from being interpreted as a part of an ill-formed for-in loop.
         * @param {ASTNode***REMOVED*** node Ancestor of an 'in' expression.
         * @param {ASTNode***REMOVED*** child Child of the node, ancestor of the same 'in' expression or the 'in' expression itself.
         * @returns {boolean***REMOVED*** True if the keyword 'in' would be interpreted as the 'in' operator, without any parenthesis.
         */
        function isSafelyEnclosingInExpression(node, child) {
            switch (node.type) {
                case "ArrayExpression":
                case "ArrayPattern":
                case "BlockStatement":
                case "ObjectExpression":
                case "ObjectPattern":
                case "TemplateLiteral":
                    return true;
                case "ArrowFunctionExpression":
                case "FunctionExpression":
                    return node.params.includes(child);
                case "CallExpression":
                case "NewExpression":
                    return node.arguments.includes(child);
                case "MemberExpression":
                    return node.computed && node.property === child;
                case "ConditionalExpression":
                    return node.consequent === child;
                default:
                    return false;
            ***REMOVED***
        ***REMOVED***

        /**
         * Starts a new reports buffering. Warnings will be stored in a buffer instead of being reported immediately.
         * An additional logic that requires multiple nodes (e.g. a whole subtree) may dismiss some of the stored warnings.
         * @returns {void***REMOVED***
         */
        function startNewReportsBuffering() {
            reportsBuffer = {
                upper: reportsBuffer,
                inExpressionNodes: [],
                reports: []
            ***REMOVED***;
        ***REMOVED***

        /**
         * Ends the current reports buffering.
         * @returns {void***REMOVED***
         */
        function endCurrentReportsBuffering() {
            const { upper, inExpressionNodes, reports ***REMOVED*** = reportsBuffer;

            if (upper) {
                upper.inExpressionNodes.push(...inExpressionNodes);
                upper.reports.push(...reports);
            ***REMOVED*** else {

                // flush remaining reports
                reports.forEach(({ finishReport ***REMOVED***) => finishReport());
            ***REMOVED***

            reportsBuffer = upper;
        ***REMOVED***

        /**
         * Checks whether the given node is in the current reports buffer.
         * @param {ASTNode***REMOVED*** node Node to check.
         * @returns {boolean***REMOVED*** True if the node is in the current buffer, false otherwise.
         */
        function isInCurrentReportsBuffer(node) {
            return reportsBuffer.reports.some(r => r.node === node);
        ***REMOVED***

        /**
         * Removes the given node from the current reports buffer.
         * @param {ASTNode***REMOVED*** node Node to remove.
         * @returns {void***REMOVED***
         */
        function removeFromCurrentReportsBuffer(node) {
            reportsBuffer.reports = reportsBuffer.reports.filter(r => r.node !== node);
        ***REMOVED***

        /**
         * Checks whether a node is a MemberExpression at NewExpression's callee.
         * @param {ASTNode***REMOVED*** node node to check.
         * @returns {boolean***REMOVED*** True if the node is a MemberExpression at NewExpression's callee. false otherwise.
         */
        function isMemberExpInNewCallee(node) {
            if (node.type === "MemberExpression") {
                return node.parent.type === "NewExpression" && node.parent.callee === node
                    ? true
                    : node.parent.object === node && isMemberExpInNewCallee(node.parent);
            ***REMOVED***
            return false;
        ***REMOVED***

        return {
            ArrayExpression(node) {
                node.elements
                    .filter(e => e && hasExcessParensWithPrecedence(e, PRECEDENCE_OF_ASSIGNMENT_EXPR))
                    .forEach(report);
            ***REMOVED***,

            ArrayPattern(node) {
                node.elements
                    .filter(e => canBeAssignmentTarget(e) && hasExcessParens(e))
                    .forEach(report);
            ***REMOVED***,

            ArrowFunctionExpression(node) {
                if (isReturnAssignException(node)) {
                    return;
                ***REMOVED***

                if (node.body.type === "ConditionalExpression" &&
                    IGNORE_ARROW_CONDITIONALS
                ) {
                    return;
                ***REMOVED***

                if (node.body.type !== "BlockStatement") {
                    const firstBodyToken = sourceCode.getFirstToken(node.body, astUtils.isNotOpeningParenToken);
                    const tokenBeforeFirst = sourceCode.getTokenBefore(firstBodyToken);

                    if (astUtils.isOpeningParenToken(tokenBeforeFirst) && astUtils.isOpeningBraceToken(firstBodyToken)) {
                        tokensToIgnore.add(firstBodyToken);
                    ***REMOVED***
                    if (hasExcessParensWithPrecedence(node.body, PRECEDENCE_OF_ASSIGNMENT_EXPR)) {
                        report(node.body);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***,

            AssignmentExpression(node) {
                if (canBeAssignmentTarget(node.left) && hasExcessParens(node.left)) {
                    report(node.left);
                ***REMOVED***

                if (!isReturnAssignException(node) && hasExcessParensWithPrecedence(node.right, precedence(node))) {
                    report(node.right);
                ***REMOVED***
            ***REMOVED***,

            BinaryExpression(node) {
                if (reportsBuffer && node.operator === "in") {
                    reportsBuffer.inExpressionNodes.push(node);
                ***REMOVED***

                checkBinaryLogical(node);
            ***REMOVED***,

            CallExpression: checkCallNew,

            ClassBody(node) {
                node.body
                    .filter(member => member.type === "MethodDefinition" && member.computed && member.key)
                    .filter(member => hasExcessParensWithPrecedence(member.key, PRECEDENCE_OF_ASSIGNMENT_EXPR))
                    .forEach(member => report(member.key));
            ***REMOVED***,

            ConditionalExpression(node) {
                if (isReturnAssignException(node)) {
                    return;
                ***REMOVED***
                if (
                    !isCondAssignException(node) &&
                    hasExcessParensWithPrecedence(node.test, precedence({ type: "LogicalExpression", operator: "||" ***REMOVED***))
                ) {
                    report(node.test);
                ***REMOVED***

                if (hasExcessParensWithPrecedence(node.consequent, PRECEDENCE_OF_ASSIGNMENT_EXPR)) {
                    report(node.consequent);
                ***REMOVED***

                if (hasExcessParensWithPrecedence(node.alternate, PRECEDENCE_OF_ASSIGNMENT_EXPR)) {
                    report(node.alternate);
                ***REMOVED***
            ***REMOVED***,

            DoWhileStatement(node) {
                if (hasExcessParens(node.test) && !isCondAssignException(node)) {
                    report(node.test);
                ***REMOVED***
            ***REMOVED***,

            ExportDefaultDeclaration: node => checkExpressionOrExportStatement(node.declaration),
            ExpressionStatement: node => checkExpressionOrExportStatement(node.expression),

            ForInStatement(node) {
                if (node.left.type !== "VariableDeclaration") {
                    const firstLeftToken = sourceCode.getFirstToken(node.left, astUtils.isNotOpeningParenToken);

                    if (
                        firstLeftToken.value === "let" &&
                        astUtils.isOpeningBracketToken(
                            sourceCode.getTokenAfter(firstLeftToken, astUtils.isNotClosingParenToken)
                        )
                    ) {

                        // ForInStatement#left expression cannot start with `let[`.
                        tokensToIgnore.add(firstLeftToken);
                    ***REMOVED***
                ***REMOVED***

                if (hasExcessParens(node.left)) {
                    report(node.left);
                ***REMOVED***

                if (hasExcessParens(node.right)) {
                    report(node.right);
                ***REMOVED***
            ***REMOVED***,

            ForOfStatement(node) {
                if (node.left.type !== "VariableDeclaration") {
                    const firstLeftToken = sourceCode.getFirstToken(node.left, astUtils.isNotOpeningParenToken);

                    if (firstLeftToken.value === "let") {

                        // ForOfStatement#left expression cannot start with `let`.
                        tokensToIgnore.add(firstLeftToken);
                    ***REMOVED***
                ***REMOVED***

                if (hasExcessParens(node.left)) {
                    report(node.left);
                ***REMOVED***

                if (hasExcessParensWithPrecedence(node.right, PRECEDENCE_OF_ASSIGNMENT_EXPR)) {
                    report(node.right);
                ***REMOVED***
            ***REMOVED***,

            ForStatement(node) {
                if (node.test && hasExcessParens(node.test) && !isCondAssignException(node)) {
                    report(node.test);
                ***REMOVED***

                if (node.update && hasExcessParens(node.update)) {
                    report(node.update);
                ***REMOVED***

                if (node.init) {

                    if (node.init.type !== "VariableDeclaration") {
                        const firstToken = sourceCode.getFirstToken(node.init, astUtils.isNotOpeningParenToken);

                        if (
                            firstToken.value === "let" &&
                            astUtils.isOpeningBracketToken(
                                sourceCode.getTokenAfter(firstToken, astUtils.isNotClosingParenToken)
                            )
                        ) {

                            // ForStatement#init expression cannot start with `let[`.
                            tokensToIgnore.add(firstToken);
                        ***REMOVED***
                    ***REMOVED***

                    startNewReportsBuffering();

                    if (hasExcessParens(node.init)) {
                        report(node.init);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***,

            "ForStatement > *.init:exit"(node) {

                /*
                 * Removing parentheses around `in` expressions might change semantics and cause errors.
                 *
                 * For example, this valid for loop:
                 *      for (let a = (b in c); ;);
                 * after removing parentheses would be treated as an invalid for-in loop:
                 *      for (let a = b in c; ;);
                 */

                if (reportsBuffer.reports.length) {
                    reportsBuffer.inExpressionNodes.forEach(inExpressionNode => {
                        const path = pathToDescendant(node, inExpressionNode);
                        let nodeToExclude;

                        for (let i = 0; i < path.length; i++) {
                            const pathNode = path[i];

                            if (i < path.length - 1) {
                                const nextPathNode = path[i + 1];

                                if (isSafelyEnclosingInExpression(pathNode, nextPathNode)) {

                                    // The 'in' expression in safely enclosed by the syntax of its ancestor nodes (e.g. by '{***REMOVED***' or '[]').
                                    return;
                                ***REMOVED***
                            ***REMOVED***

                            if (isParenthesised(pathNode)) {
                                if (isInCurrentReportsBuffer(pathNode)) {

                                    // This node was supposed to be reported, but parentheses might be necessary.

                                    if (isParenthesisedTwice(pathNode)) {

                                        /*
                                         * This node is parenthesised twice, it certainly has at least one pair of `extra` parentheses.
                                         * If the --fix option is on, the current fixing iteration will remove only one pair of parentheses.
                                         * The remaining pair is safely enclosing the 'in' expression.
                                         */
                                        return;
                                    ***REMOVED***

                                    // Exclude the outermost node only.
                                    if (!nodeToExclude) {
                                        nodeToExclude = pathNode;
                                    ***REMOVED***

                                    // Don't break the loop here, there might be some safe nodes or parentheses that will stay inside.

                                ***REMOVED*** else {

                                    // This node will stay parenthesised, the 'in' expression in safely enclosed by '()'.
                                    return;
                                ***REMOVED***
                            ***REMOVED***
                        ***REMOVED***

                        // Exclude the node from the list (i.e. treat parentheses as necessary)
                        removeFromCurrentReportsBuffer(nodeToExclude);
                    ***REMOVED***);
                ***REMOVED***

                endCurrentReportsBuffering();
            ***REMOVED***,

            IfStatement(node) {
                if (hasExcessParens(node.test) && !isCondAssignException(node)) {
                    report(node.test);
                ***REMOVED***
            ***REMOVED***,

            ImportExpression(node) {
                const { source ***REMOVED*** = node;

                if (source.type === "SequenceExpression") {
                    if (hasDoubleExcessParens(source)) {
                        report(source);
                    ***REMOVED***
                ***REMOVED*** else if (hasExcessParens(source)) {
                    report(source);
                ***REMOVED***
            ***REMOVED***,

            LogicalExpression: checkBinaryLogical,

            MemberExpression(node) {
                const shouldAllowWrapOnce = isMemberExpInNewCallee(node) &&
                  doesMemberExpressionContainCallExpression(node);
                const nodeObjHasExcessParens = shouldAllowWrapOnce
                    ? hasDoubleExcessParens(node.object)
                    : hasExcessParens(node.object) &&
                    !(
                        isImmediateFunctionPrototypeMethodCall(node.parent) &&
                        node.parent.callee === node &&
                        IGNORE_FUNCTION_PROTOTYPE_METHODS
                    );

                if (
                    nodeObjHasExcessParens &&
                    precedence(node.object) >= precedence(node) &&
                    (
                        node.computed ||
                        !(
                            astUtils.isDecimalInteger(node.object) ||

                            // RegExp literal is allowed to have parens (#1589)
                            (node.object.type === "Literal" && node.object.regex)
                        )
                    )
                ) {
                    report(node.object);
                ***REMOVED***

                if (nodeObjHasExcessParens &&
                  node.object.type === "CallExpression"
                ) {
                    report(node.object);
                ***REMOVED***

                if (nodeObjHasExcessParens &&
                  !IGNORE_NEW_IN_MEMBER_EXPR &&
                  node.object.type === "NewExpression" &&
                  isNewExpressionWithParens(node.object)) {
                    report(node.object);
                ***REMOVED***

                if (nodeObjHasExcessParens &&
                    node.optional &&
                    node.object.type === "ChainExpression"
                ) {
                    report(node.object);
                ***REMOVED***

                if (node.computed && hasExcessParens(node.property)) {
                    report(node.property);
                ***REMOVED***
            ***REMOVED***,

            NewExpression: checkCallNew,

            ObjectExpression(node) {
                node.properties
                    .filter(property => property.value && hasExcessParensWithPrecedence(property.value, PRECEDENCE_OF_ASSIGNMENT_EXPR))
                    .forEach(property => report(property.value));
            ***REMOVED***,

            ObjectPattern(node) {
                node.properties
                    .filter(property => {
                        const value = property.value;

                        return canBeAssignmentTarget(value) && hasExcessParens(value);
                    ***REMOVED***).forEach(property => report(property.value));
            ***REMOVED***,

            Property(node) {
                if (node.computed) {
                    const { key ***REMOVED*** = node;

                    if (key && hasExcessParensWithPrecedence(key, PRECEDENCE_OF_ASSIGNMENT_EXPR)) {
                        report(key);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***,

            RestElement(node) {
                const argument = node.argument;

                if (canBeAssignmentTarget(argument) && hasExcessParens(argument)) {
                    report(argument);
                ***REMOVED***
            ***REMOVED***,

            ReturnStatement(node) {
                const returnToken = sourceCode.getFirstToken(node);

                if (isReturnAssignException(node)) {
                    return;
                ***REMOVED***

                if (node.argument &&
                        hasExcessParensNoLineTerminator(returnToken, node.argument) &&

                        // RegExp literal is allowed to have parens (#1589)
                        !(node.argument.type === "Literal" && node.argument.regex)) {
                    report(node.argument);
                ***REMOVED***
            ***REMOVED***,

            SequenceExpression(node) {
                const precedenceOfNode = precedence(node);

                node.expressions
                    .filter(e => hasExcessParensWithPrecedence(e, precedenceOfNode))
                    .forEach(report);
            ***REMOVED***,

            SwitchCase(node) {
                if (node.test && hasExcessParens(node.test)) {
                    report(node.test);
                ***REMOVED***
            ***REMOVED***,

            SwitchStatement(node) {
                if (hasExcessParens(node.discriminant)) {
                    report(node.discriminant);
                ***REMOVED***
            ***REMOVED***,

            ThrowStatement(node) {
                const throwToken = sourceCode.getFirstToken(node);

                if (hasExcessParensNoLineTerminator(throwToken, node.argument)) {
                    report(node.argument);
                ***REMOVED***
            ***REMOVED***,

            UnaryExpression: checkArgumentWithPrecedence,
            UpdateExpression(node) {
                if (node.prefix) {
                    checkArgumentWithPrecedence(node);
                ***REMOVED*** else {
                    const { argument ***REMOVED*** = node;
                    const operatorToken = sourceCode.getLastToken(node);

                    if (argument.loc.end.line === operatorToken.loc.start.line) {
                        checkArgumentWithPrecedence(node);
                    ***REMOVED*** else {
                        if (hasDoubleExcessParens(argument)) {
                            report(argument);
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***,
            AwaitExpression: checkArgumentWithPrecedence,

            VariableDeclarator(node) {
                if (
                    node.init && hasExcessParensWithPrecedence(node.init, PRECEDENCE_OF_ASSIGNMENT_EXPR) &&

                    // RegExp literal is allowed to have parens (#1589)
                    !(node.init.type === "Literal" && node.init.regex)
                ) {
                    report(node.init);
                ***REMOVED***
            ***REMOVED***,

            WhileStatement(node) {
                if (hasExcessParens(node.test) && !isCondAssignException(node)) {
                    report(node.test);
                ***REMOVED***
            ***REMOVED***,

            WithStatement(node) {
                if (hasExcessParens(node.object)) {
                    report(node.object);
                ***REMOVED***
            ***REMOVED***,

            YieldExpression(node) {
                if (node.argument) {
                    const yieldToken = sourceCode.getFirstToken(node);

                    if ((precedence(node.argument) >= precedence(node) &&
                            hasExcessParensNoLineTerminator(yieldToken, node.argument)) ||
                            hasDoubleExcessParens(node.argument)) {
                        report(node.argument);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***,

            ClassDeclaration: checkClass,
            ClassExpression: checkClass,

            SpreadElement: checkSpreadOperator,
            SpreadProperty: checkSpreadOperator,
            ExperimentalSpreadProperty: checkSpreadOperator,

            TemplateLiteral(node) {
                node.expressions
                    .filter(e => e && hasExcessParens(e))
                    .forEach(report);
            ***REMOVED***,

            AssignmentPattern(node) {
                const { left, right ***REMOVED*** = node;

                if (canBeAssignmentTarget(left) && hasExcessParens(left)) {
                    report(left);
                ***REMOVED***

                if (right && hasExcessParensWithPrecedence(right, PRECEDENCE_OF_ASSIGNMENT_EXPR)) {
                    report(right);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
