// eslint/lib/rules/no-fallthrough.js
/**
 * @fileoverview Rule to flag fall-through cases in switch statements.
 * @author Matt DuVall <http://mattduvall.com/>
 */
"use strict";

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

const DEFAULT_FALLTHROUGH_COMMENT = /falls?\s?through/iu;

/**
 * Checks whether or not a given case has a fallthrough comment.
 * @param {ASTNode***REMOVED*** caseWhichFallsThrough SwitchCase node which falls through.
 * @param {ASTNode***REMOVED*** subsequentCase The case after caseWhichFallsThrough.
 * @param {RuleContext***REMOVED*** context A rule context which stores comments.
 * @param {RegExp***REMOVED*** fallthroughCommentPattern A pattern to match comment to.
 * @returns {boolean***REMOVED*** `true` if the case has a valid fallthrough comment.
 */
function hasFallthroughComment(caseWhichFallsThrough, subsequentCase, context, fallthroughCommentPattern) {
    const sourceCode = context.getSourceCode();

    if (caseWhichFallsThrough.consequent.length === 1 && caseWhichFallsThrough.consequent[0].type === "BlockStatement") {
        const trailingCloseBrace = sourceCode.getLastToken(caseWhichFallsThrough.consequent[0]);
        const commentInBlock = sourceCode.getCommentsBefore(trailingCloseBrace).pop();

        if (commentInBlock && fallthroughCommentPattern.test(commentInBlock.value)) {
            return true;
        ***REMOVED***
    ***REMOVED***

    const comment = sourceCode.getCommentsBefore(subsequentCase).pop();

    return Boolean(comment && fallthroughCommentPattern.test(comment.value));
***REMOVED***

/**
 * Checks whether or not a given code path segment is reachable.
 * @param {CodePathSegment***REMOVED*** segment A CodePathSegment to check.
 * @returns {boolean***REMOVED*** `true` if the segment is reachable.
 */
function isReachable(segment) {
    return segment.reachable;
***REMOVED***

/**
 * Checks whether a node and a token are separated by blank lines
 * @param {ASTNode***REMOVED*** node The node to check
 * @param {Token***REMOVED*** token The token to compare against
 * @returns {boolean***REMOVED*** `true` if there are blank lines between node and token
 */
function hasBlankLinesBetween(node, token) {
    return token.loc.start.line > node.loc.end.line + 1;
***REMOVED***

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "problem",

        docs: {
            description: "disallow fallthrough of `case` statements",
            category: "Best Practices",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-fallthrough"
        ***REMOVED***,

        schema: [
            {
                type: "object",
                properties: {
                    commentPattern: {
                        type: "string",
                        default: ""
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],
        messages: {
            case: "Expected a 'break' statement before 'case'.",
            default: "Expected a 'break' statement before 'default'."
        ***REMOVED***
    ***REMOVED***,

    create(context) {
        const options = context.options[0] || {***REMOVED***;
        let currentCodePath = null;
        const sourceCode = context.getSourceCode();

        /*
         * We need to use leading comments of the next SwitchCase node because
         * trailing comments is wrong if semicolons are omitted.
         */
        let fallthroughCase = null;
        let fallthroughCommentPattern = null;

        if (options.commentPattern) {
            fallthroughCommentPattern = new RegExp(options.commentPattern, "u");
        ***REMOVED*** else {
            fallthroughCommentPattern = DEFAULT_FALLTHROUGH_COMMENT;
        ***REMOVED***

        return {
            onCodePathStart(codePath) {
                currentCodePath = codePath;
            ***REMOVED***,
            onCodePathEnd() {
                currentCodePath = currentCodePath.upper;
            ***REMOVED***,

            SwitchCase(node) {

                /*
                 * Checks whether or not there is a fallthrough comment.
                 * And reports the previous fallthrough node if that does not exist.
                 */
                if (fallthroughCase && !hasFallthroughComment(fallthroughCase, node, context, fallthroughCommentPattern)) {
                    context.report({
                        messageId: node.test ? "case" : "default",
                        node
                    ***REMOVED***);
                ***REMOVED***
                fallthroughCase = null;
            ***REMOVED***,

            "SwitchCase:exit"(node) {
                const nextToken = sourceCode.getTokenAfter(node);

                /*
                 * `reachable` meant fall through because statements preceded by
                 * `break`, `return`, or `throw` are unreachable.
                 * And allows empty cases and the last case.
                 */
                if (currentCodePath.currentSegments.some(isReachable) &&
                    (node.consequent.length > 0 || hasBlankLinesBetween(node, nextToken)) &&
                    node.parent.cases[node.parent.cases.length - 1] !== node) {
                    fallthroughCase = node;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
