// eslint/lib/rules/no-func-assign.js
/**
 * @fileoverview Rule to flag use of function declaration identifiers as variables.
 * @author Ian Christian Myers
 */

"use strict";

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "problem",

        docs: {
            description: "disallow reassigning `function` declarations",
            category: "Possible Errors",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-func-assign"
        ***REMOVED***,

        schema: [],

        messages: {
            isAFunction: "'{{name***REMOVED******REMOVED***' is a function."
        ***REMOVED***
    ***REMOVED***,

    create(context) {

        /**
         * Reports a reference if is non initializer and writable.
         * @param {References***REMOVED*** references Collection of reference to check.
         * @returns {void***REMOVED***
         */
        function checkReference(references) {
            astUtils.getModifyingReferences(references).forEach(reference => {
                context.report({
                    node: reference.identifier,
                    messageId: "isAFunction",
                    data: {
                        name: reference.identifier.name
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***

        /**
         * Finds and reports references that are non initializer and writable.
         * @param {Variable***REMOVED*** variable A variable to check.
         * @returns {void***REMOVED***
         */
        function checkVariable(variable) {
            if (variable.defs[0].type === "FunctionName") {
                checkReference(variable.references);
            ***REMOVED***
        ***REMOVED***

        /**
         * Checks parameters of a given function node.
         * @param {ASTNode***REMOVED*** node A function node to check.
         * @returns {void***REMOVED***
         */
        function checkForFunction(node) {
            context.getDeclaredVariables(node).forEach(checkVariable);
        ***REMOVED***

        return {
            FunctionDeclaration: checkForFunction,
            FunctionExpression: checkForFunction
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
