// eslint/lib/rules/no-promise-executor-return.js
/**
 * @fileoverview Rule to disallow returning values from Promise executor functions
 * @author Milos Djermanovic
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const { findVariable ***REMOVED*** = require("eslint-utils");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

const functionTypesToCheck = new Set(["ArrowFunctionExpression", "FunctionExpression"]);

/**
 * Determines whether the given identifier node is a reference to a global variable.
 * @param {ASTNode***REMOVED*** node `Identifier` node to check.
 * @param {Scope***REMOVED*** scope Scope to which the node belongs.
 * @returns {boolean***REMOVED*** True if the identifier is a reference to a global variable.
 */
function isGlobalReference(node, scope) {
    const variable = findVariable(scope, node);

    return variable !== null && variable.scope.type === "global" && variable.defs.length === 0;
***REMOVED***

/**
 * Finds function's outer scope.
 * @param {Scope***REMOVED*** scope Function's own scope.
 * @returns {Scope***REMOVED*** Function's outer scope.
 */
function getOuterScope(scope) {
    const upper = scope.upper;

    if (upper.type === "function-expression-name") {
        return upper.upper;
    ***REMOVED***
    return upper;
***REMOVED***

/**
 * Determines whether the given function node is used as a Promise executor.
 * @param {ASTNode***REMOVED*** node The node to check.
 * @param {Scope***REMOVED*** scope Function's own scope.
 * @returns {boolean***REMOVED*** `true` if the node is a Promise executor.
 */
function isPromiseExecutor(node, scope) {
    const parent = node.parent;

    return parent.type === "NewExpression" &&
        parent.arguments[0] === node &&
        parent.callee.type === "Identifier" &&
        parent.callee.name === "Promise" &&
        isGlobalReference(parent.callee, getOuterScope(scope));
***REMOVED***

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "problem",

        docs: {
            description: "disallow returning values from Promise executor functions",
            category: "Possible Errors",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-promise-executor-return"
        ***REMOVED***,

        schema: [],

        messages: {
            returnsValue: "Return values from promise executor functions cannot be read."
        ***REMOVED***
    ***REMOVED***,

    create(context) {

        let funcInfo = null;

        /**
         * Reports the given node.
         * @param {ASTNode***REMOVED*** node Node to report.
         * @returns {void***REMOVED***
         */
        function report(node) {
            context.report({ node, messageId: "returnsValue" ***REMOVED***);
        ***REMOVED***

        return {

            onCodePathStart(_, node) {
                funcInfo = {
                    upper: funcInfo,
                    shouldCheck: functionTypesToCheck.has(node.type) && isPromiseExecutor(node, context.getScope())
                ***REMOVED***;

                if (funcInfo.shouldCheck && node.type === "ArrowFunctionExpression" && node.expression) {
                    report(node.body);
                ***REMOVED***
            ***REMOVED***,

            onCodePathEnd() {
                funcInfo = funcInfo.upper;
            ***REMOVED***,

            ReturnStatement(node) {
                if (funcInfo.shouldCheck && node.argument) {
                    report(node);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
