// eslint/lib/rules/no-param-reassign.js
/**
 * @fileoverview Disallow reassignment of function parameters.
 * @author Nat Burns
 */
"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

const stopNodePattern = /(?:Statement|Declaration|Function(?:Expression)?|Program)$/u;

module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "disallow reassigning `function` parameters",
            category: "Best Practices",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-param-reassign"
        ***REMOVED***,

        schema: [
            {
                oneOf: [
                    {
                        type: "object",
                        properties: {
                            props: {
                                enum: [false]
                            ***REMOVED***
                        ***REMOVED***,
                        additionalProperties: false
                    ***REMOVED***,
                    {
                        type: "object",
                        properties: {
                            props: {
                                enum: [true]
                            ***REMOVED***,
                            ignorePropertyModificationsFor: {
                                type: "array",
                                items: {
                                    type: "string"
                                ***REMOVED***,
                                uniqueItems: true
                            ***REMOVED***,
                            ignorePropertyModificationsForRegex: {
                                type: "array",
                                items: {
                                    type: "string"
                                ***REMOVED***,
                                uniqueItems: true
                            ***REMOVED***
                        ***REMOVED***,
                        additionalProperties: false
                    ***REMOVED***
                ]
            ***REMOVED***
        ],

        messages: {
            assignmentToFunctionParam: "Assignment to function parameter '{{name***REMOVED******REMOVED***'.",
            assignmentToFunctionParamProp: "Assignment to property of function parameter '{{name***REMOVED******REMOVED***'."
        ***REMOVED***
    ***REMOVED***,

    create(context) {
        const props = context.options[0] && context.options[0].props;
        const ignoredPropertyAssignmentsFor = context.options[0] && context.options[0].ignorePropertyModificationsFor || [];
        const ignoredPropertyAssignmentsForRegex = context.options[0] && context.options[0].ignorePropertyModificationsForRegex || [];

        /**
         * Checks whether or not the reference modifies properties of its variable.
         * @param {Reference***REMOVED*** reference A reference to check.
         * @returns {boolean***REMOVED*** Whether or not the reference modifies properties of its variable.
         */
        function isModifyingProp(reference) {
            let node = reference.identifier;
            let parent = node.parent;

            while (parent && (!stopNodePattern.test(parent.type) ||
                    parent.type === "ForInStatement" || parent.type === "ForOfStatement")) {
                switch (parent.type) {

                    // e.g. foo.a = 0;
                    case "AssignmentExpression":
                        return parent.left === node;

                    // e.g. ++foo.a;
                    case "UpdateExpression":
                        return true;

                    // e.g. delete foo.a;
                    case "UnaryExpression":
                        if (parent.operator === "delete") {
                            return true;
                        ***REMOVED***
                        break;

                    // e.g. for (foo.a in b) {***REMOVED***
                    case "ForInStatement":
                    case "ForOfStatement":
                        if (parent.left === node) {
                            return true;
                        ***REMOVED***

                        // this is a stop node for parent.right and parent.body
                        return false;

                    // EXCLUDES: e.g. cache.get(foo.a).b = 0;
                    case "CallExpression":
                        if (parent.callee !== node) {
                            return false;
                        ***REMOVED***
                        break;

                    // EXCLUDES: e.g. cache[foo.a] = 0;
                    case "MemberExpression":
                        if (parent.property === node) {
                            return false;
                        ***REMOVED***
                        break;

                    // EXCLUDES: e.g. ({ [foo]: a ***REMOVED***) = bar;
                    case "Property":
                        if (parent.key === node) {
                            return false;
                        ***REMOVED***

                        break;

                    // EXCLUDES: e.g. (foo ? a : b).c = bar;
                    case "ConditionalExpression":
                        if (parent.test === node) {
                            return false;
                        ***REMOVED***

                        break;

                    // no default
                ***REMOVED***

                node = parent;
                parent = node.parent;
            ***REMOVED***

            return false;
        ***REMOVED***

        /**
         * Tests that an identifier name matches any of the ignored property assignments.
         * First we test strings in ignoredPropertyAssignmentsFor.
         * Then we instantiate and test RegExp objects from ignoredPropertyAssignmentsForRegex strings.
         * @param {string***REMOVED*** identifierName A string that describes the name of an identifier to
         * ignore property assignments for.
         * @returns {boolean***REMOVED*** Whether the string matches an ignored property assignment regular expression or not.
         */
        function isIgnoredPropertyAssignment(identifierName) {
            return ignoredPropertyAssignmentsFor.includes(identifierName) ||
                ignoredPropertyAssignmentsForRegex.some(ignored => new RegExp(ignored, "u").test(identifierName));
        ***REMOVED***

        /**
         * Reports a reference if is non initializer and writable.
         * @param {Reference***REMOVED*** reference A reference to check.
         * @param {int***REMOVED*** index The index of the reference in the references.
         * @param {Reference[]***REMOVED*** references The array that the reference belongs to.
         * @returns {void***REMOVED***
         */
        function checkReference(reference, index, references) {
            const identifier = reference.identifier;

            if (identifier &&
                !reference.init &&

                /*
                 * Destructuring assignments can have multiple default value,
                 * so possibly there are multiple writeable references for the same identifier.
                 */
                (index === 0 || references[index - 1].identifier !== identifier)
            ) {
                if (reference.isWrite()) {
                    context.report({
                        node: identifier,
                        messageId: "assignmentToFunctionParam",
                        data: { name: identifier.name ***REMOVED***
                    ***REMOVED***);
                ***REMOVED*** else if (props && isModifyingProp(reference) && !isIgnoredPropertyAssignment(identifier.name)) {
                    context.report({
                        node: identifier,
                        messageId: "assignmentToFunctionParamProp",
                        data: { name: identifier.name ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        /**
         * Finds and reports references that are non initializer and writable.
         * @param {Variable***REMOVED*** variable A variable to check.
         * @returns {void***REMOVED***
         */
        function checkVariable(variable) {
            if (variable.defs[0].type === "Parameter") {
                variable.references.forEach(checkReference);
            ***REMOVED***
        ***REMOVED***

        /**
         * Checks parameters of a given function node.
         * @param {ASTNode***REMOVED*** node A function node to check.
         * @returns {void***REMOVED***
         */
        function checkForFunction(node) {
            context.getDeclaredVariables(node).forEach(checkVariable);
        ***REMOVED***

        return {

            // `:exit` is needed for the `node.parent` property of identifier nodes.
            "FunctionDeclaration:exit": checkForFunction,
            "FunctionExpression:exit": checkForFunction,
            "ArrowFunctionExpression:exit": checkForFunction
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
