// eslint/lib/rules/no-restricted-syntax.js
/**
 * @fileoverview Rule to flag use of certain node types
 * @author Burak Yigit Kaya
 */
"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "disallow specified syntax",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-restricted-syntax"
        ***REMOVED***,

        schema: {
            type: "array",
            items: {
                oneOf: [
                    {
                        type: "string"
                    ***REMOVED***,
                    {
                        type: "object",
                        properties: {
                            selector: { type: "string" ***REMOVED***,
                            message: { type: "string" ***REMOVED***
                        ***REMOVED***,
                        required: ["selector"],
                        additionalProperties: false
                    ***REMOVED***
                ]
            ***REMOVED***,
            uniqueItems: true,
            minItems: 0
        ***REMOVED***,

        messages: {
            // eslint-disable-next-line eslint-plugin/report-message-format
            restrictedSyntax: "{{message***REMOVED******REMOVED***"
        ***REMOVED***
    ***REMOVED***,

    create(context) {
        return context.options.reduce((result, selectorOrObject) => {
            const isStringFormat = (typeof selectorOrObject === "string");
            const hasCustomMessage = !isStringFormat && Boolean(selectorOrObject.message);

            const selector = isStringFormat ? selectorOrObject : selectorOrObject.selector;
            const message = hasCustomMessage ? selectorOrObject.message : `Using '${selector***REMOVED***' is not allowed.`;

            return Object.assign(result, {
                [selector](node) {
                    context.report({
                        node,
                        messageId: "restrictedSyntax",
                        data: { message ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***, {***REMOVED***);

    ***REMOVED***
***REMOVED***;
