// eslint/lib/rules/no-useless-computed-key.js
/**
 * @fileoverview Rule to disallow unnecessary computed property keys in object literals
 * @author Burak Yigit Kaya
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Determines whether the computed key syntax is unnecessarily used for the given node.
 * In particular, it determines whether removing the square brackets and using the content between them
 * directly as the key (e.g. ['foo'] -> 'foo') would produce valid syntax and preserve the same behavior.
 * Valid non-computed keys are only: identifiers, number literals and string literals.
 * Only literals can preserve the same behavior, with a few exceptions for specific node types:
 * Property
 *   - { ["__proto__"]: foo ***REMOVED*** defines a property named "__proto__"
 *     { "__proto__": foo ***REMOVED*** defines object's prototype
 * PropertyDefinition
 *   - class C { ["constructor"]; ***REMOVED*** defines an instance field named "constructor"
 *     class C { "constructor"; ***REMOVED*** produces a parsing error
 *   - class C { static ["constructor"]; ***REMOVED*** defines a static field named "constructor"
 *     class C { static "constructor"; ***REMOVED*** produces a parsing error
 *   - class C { static ["prototype"]; ***REMOVED*** produces a runtime error (doesn't break the whole script)
 *     class C { static "prototype"; ***REMOVED*** produces a parsing error (breaks the whole script)
 * MethodDefinition
 *   - class C { ["constructor"]() {***REMOVED*** ***REMOVED*** defines a prototype method named "constructor"
 *     class C { "constructor"() {***REMOVED*** ***REMOVED*** defines the constructor
 *   - class C { static ["prototype"]() {***REMOVED*** ***REMOVED*** produces a runtime error (doesn't break the whole script)
 *     class C { static "prototype"() {***REMOVED*** ***REMOVED*** produces a parsing error (breaks the whole script)
 * @param {ASTNode***REMOVED*** node The node to check. It can be `Property`, `PropertyDefinition` or `MethodDefinition`.
 * @throws {Error***REMOVED*** (Unreachable.)
 * @returns {void***REMOVED*** `true` if the node has useless computed key.
 */
function hasUselessComputedKey(node) {
    if (!node.computed) {
        return false;
    ***REMOVED***

    const { key ***REMOVED*** = node;

    if (key.type !== "Literal") {
        return false;
    ***REMOVED***

    const { value ***REMOVED*** = key;

    if (typeof value !== "number" && typeof value !== "string") {
        return false;
    ***REMOVED***

    switch (node.type) {
        case "Property":
            return value !== "__proto__";

        case "PropertyDefinition":
            if (node.static) {
                return value !== "constructor" && value !== "prototype";
            ***REMOVED***

            return value !== "constructor";

        case "MethodDefinition":
            if (node.static) {
                return value !== "prototype";
            ***REMOVED***

            return value !== "constructor";

        /* c8 ignore next */
        default:
            throw new Error(`Unexpected node type: ${node.type***REMOVED***`);
    ***REMOVED***

***REMOVED***

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

/** @type {import('../shared/types').Rule***REMOVED*** */
module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "Disallow unnecessary computed property keys in objects and classes",
            recommended: false,
            url: "https://eslint.org/docs/latest/rules/no-useless-computed-key"
        ***REMOVED***,

        schema: [{
            type: "object",
            properties: {
                enforceForClassMembers: {
                    type: "boolean",
                    default: false
                ***REMOVED***
            ***REMOVED***,
            additionalProperties: false
        ***REMOVED***],
        fixable: "code",

        messages: {
            unnecessarilyComputedProperty: "Unnecessarily computed property [{{property***REMOVED******REMOVED***] found."
        ***REMOVED***
    ***REMOVED***,
    create(context) {
        const sourceCode = context.sourceCode;
        const enforceForClassMembers = context.options[0] && context.options[0].enforceForClassMembers;

        /**
         * Reports a given node if it violated this rule.
         * @param {ASTNode***REMOVED*** node The node to check.
         * @returns {void***REMOVED***
         */
        function check(node) {
            if (hasUselessComputedKey(node)) {
                const { key ***REMOVED*** = node;

                context.report({
                    node,
                    messageId: "unnecessarilyComputedProperty",
                    data: { property: sourceCode.getText(key) ***REMOVED***,
                    fix(fixer) {
                        const leftSquareBracket = sourceCode.getTokenBefore(key, astUtils.isOpeningBracketToken);
                        const rightSquareBracket = sourceCode.getTokenAfter(key, astUtils.isClosingBracketToken);

                        // If there are comments between the brackets and the property name, don't do a fix.
                        if (sourceCode.commentsExistBetween(leftSquareBracket, rightSquareBracket)) {
                            return null;
                        ***REMOVED***

                        const tokenBeforeLeftBracket = sourceCode.getTokenBefore(leftSquareBracket);

                        // Insert a space before the key to avoid changing identifiers, e.g. ({ get[2]() {***REMOVED*** ***REMOVED***) to ({ get2() {***REMOVED*** ***REMOVED***)
                        const needsSpaceBeforeKey = tokenBeforeLeftBracket.range[1] === leftSquareBracket.range[0] &&
                            !astUtils.canTokensBeAdjacent(tokenBeforeLeftBracket, sourceCode.getFirstToken(key));

                        const replacementKey = (needsSpaceBeforeKey ? " " : "") + key.raw;

                        return fixer.replaceTextRange([leftSquareBracket.range[0], rightSquareBracket.range[1]], replacementKey);
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        /**
         * A no-op function to act as placeholder for checking a node when the `enforceForClassMembers` option is `false`.
         * @returns {void***REMOVED***
         * @private
         */
        function noop() {***REMOVED***

        return {
            Property: check,
            MethodDefinition: enforceForClassMembers ? check : noop,
            PropertyDefinition: enforceForClassMembers ? check : noop
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
