// eslint/lib/rules/no-unused-expressions.js
/**
 * @fileoverview Flag expressions in statement position that do not side effect
 * @author Michael Ficarra
 */
"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

/**
 * Returns `true`.
 * @returns {boolean***REMOVED*** `true`.
 */
function alwaysTrue() {
    return true;
***REMOVED***

/**
 * Returns `false`.
 * @returns {boolean***REMOVED*** `false`.
 */
function alwaysFalse() {
    return false;
***REMOVED***

module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "disallow unused expressions",
            category: "Best Practices",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-unused-expressions"
        ***REMOVED***,

        schema: [
            {
                type: "object",
                properties: {
                    allowShortCircuit: {
                        type: "boolean",
                        default: false
                    ***REMOVED***,
                    allowTernary: {
                        type: "boolean",
                        default: false
                    ***REMOVED***,
                    allowTaggedTemplates: {
                        type: "boolean",
                        default: false
                    ***REMOVED***,
                    enforceForJSX: {
                        type: "boolean",
                        default: false
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],

        messages: {
            unusedExpression: "Expected an assignment or function call and instead saw an expression."
        ***REMOVED***
    ***REMOVED***,

    create(context) {
        const config = context.options[0] || {***REMOVED***,
            allowShortCircuit = config.allowShortCircuit || false,
            allowTernary = config.allowTernary || false,
            allowTaggedTemplates = config.allowTaggedTemplates || false,
            enforceForJSX = config.enforceForJSX || false;

        // eslint-disable-next-line jsdoc/require-description
        /**
         * @param {ASTNode***REMOVED*** node any node
         * @returns {boolean***REMOVED*** whether the given node structurally represents a directive
         */
        function looksLikeDirective(node) {
            return node.type === "ExpressionStatement" &&
                node.expression.type === "Literal" && typeof node.expression.value === "string";
        ***REMOVED***

        // eslint-disable-next-line jsdoc/require-description
        /**
         * @param {Function***REMOVED*** predicate ([a] -> Boolean) the function used to make the determination
         * @param {a[]***REMOVED*** list the input list
         * @returns {a[]***REMOVED*** the leading sequence of members in the given list that pass the given predicate
         */
        function takeWhile(predicate, list) {
            for (let i = 0; i < list.length; ++i) {
                if (!predicate(list[i])) {
                    return list.slice(0, i);
                ***REMOVED***
            ***REMOVED***
            return list.slice();
        ***REMOVED***

        // eslint-disable-next-line jsdoc/require-description
        /**
         * @param {ASTNode***REMOVED*** node a Program or BlockStatement node
         * @returns {ASTNode[]***REMOVED*** the leading sequence of directive nodes in the given node's body
         */
        function directives(node) {
            return takeWhile(looksLikeDirective, node.body);
        ***REMOVED***

        // eslint-disable-next-line jsdoc/require-description
        /**
         * @param {ASTNode***REMOVED*** node any node
         * @param {ASTNode[]***REMOVED*** ancestors the given node's ancestors
         * @returns {boolean***REMOVED*** whether the given node is considered a directive in its current position
         */
        function isDirective(node, ancestors) {
            const parent = ancestors[ancestors.length - 1],
                grandparent = ancestors[ancestors.length - 2];

            return (parent.type === "Program" || parent.type === "BlockStatement" &&
                    (/Function/u.test(grandparent.type))) &&
                    directives(parent).indexOf(node) >= 0;
        ***REMOVED***

        /**
         * The member functions return `true` if the type has no side-effects.
         * Unknown nodes are handled as `false`, then this rule ignores those.
         */
        const Checker = Object.assign(Object.create(null), {
            isDisallowed(node) {
                return (Checker[node.type] || alwaysFalse)(node);
            ***REMOVED***,

            ArrayExpression: alwaysTrue,
            ArrowFunctionExpression: alwaysTrue,
            BinaryExpression: alwaysTrue,
            ChainExpression(node) {
                return Checker.isDisallowed(node.expression);
            ***REMOVED***,
            ClassExpression: alwaysTrue,
            ConditionalExpression(node) {
                if (allowTernary) {
                    return Checker.isDisallowed(node.consequent) || Checker.isDisallowed(node.alternate);
                ***REMOVED***
                return true;
            ***REMOVED***,
            FunctionExpression: alwaysTrue,
            Identifier: alwaysTrue,
            JSXElement() {
                return enforceForJSX;
            ***REMOVED***,
            JSXFragment() {
                return enforceForJSX;
            ***REMOVED***,
            Literal: alwaysTrue,
            LogicalExpression(node) {
                if (allowShortCircuit) {
                    return Checker.isDisallowed(node.right);
                ***REMOVED***
                return true;
            ***REMOVED***,
            MemberExpression: alwaysTrue,
            MetaProperty: alwaysTrue,
            ObjectExpression: alwaysTrue,
            SequenceExpression: alwaysTrue,
            TaggedTemplateExpression() {
                return !allowTaggedTemplates;
            ***REMOVED***,
            TemplateLiteral: alwaysTrue,
            ThisExpression: alwaysTrue,
            UnaryExpression(node) {
                return node.operator !== "void" && node.operator !== "delete";
            ***REMOVED***
        ***REMOVED***);

        return {
            ExpressionStatement(node) {
                if (Checker.isDisallowed(node.expression) && !isDirective(node, context.getAncestors())) {
                    context.report({ node, messageId: "unusedExpression" ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
