// eslint/lib/rules/no-unused-vars.js
/**
 * @fileoverview Rule to flag declared but unused variables
 * @author Ilya Volodin
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Typedefs
//------------------------------------------------------------------------------

/**
 * A simple name for the types of variables that this rule supports
 * @typedef {'array-destructure'|'catch-clause'|'parameter'|'variable'***REMOVED*** VariableType
 */

/**
 * Bag of data used for formatting the `unusedVar` lint message.
 * @typedef {Object***REMOVED*** UnusedVarMessageData
 * @property {string***REMOVED*** varName The name of the unused var.
 * @property {'defined'|'assigned a value'***REMOVED*** action Description of the vars state.
 * @property {string***REMOVED*** additional Any additional info to be appended at the end.
 */

/**
 * Bag of data used for formatting the `usedIgnoredVar` lint message.
 * @typedef {Object***REMOVED*** UsedIgnoredVarMessageData
 * @property {string***REMOVED*** varName The name of the unused var.
 * @property {string***REMOVED*** additional Any additional info to be appended at the end.
 */

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

/** @type {import('../types').Rule.RuleModule***REMOVED*** */
module.exports = {
	meta: {
		type: "problem",

		docs: {
			description: "Disallow unused variables",
			recommended: true,
			url: "https://eslint.org/docs/latest/rules/no-unused-vars",
		***REMOVED***,

		hasSuggestions: true,

		schema: [
			{
				oneOf: [
					{
						enum: ["all", "local"],
					***REMOVED***,
					{
						type: "object",
						properties: {
							vars: {
								enum: ["all", "local"],
							***REMOVED***,
							varsIgnorePattern: {
								type: "string",
							***REMOVED***,
							args: {
								enum: ["all", "after-used", "none"],
							***REMOVED***,
							ignoreRestSiblings: {
								type: "boolean",
							***REMOVED***,
							argsIgnorePattern: {
								type: "string",
							***REMOVED***,
							caughtErrors: {
								enum: ["all", "none"],
							***REMOVED***,
							caughtErrorsIgnorePattern: {
								type: "string",
							***REMOVED***,
							destructuredArrayIgnorePattern: {
								type: "string",
							***REMOVED***,
							ignoreClassWithStaticInitBlock: {
								type: "boolean",
							***REMOVED***,
							reportUsedIgnorePattern: {
								type: "boolean",
							***REMOVED***,
						***REMOVED***,
						additionalProperties: false,
					***REMOVED***,
				],
			***REMOVED***,
		],

		messages: {
			unusedVar:
				"'{{varName***REMOVED******REMOVED***' is {{action***REMOVED******REMOVED*** but never used{{additional***REMOVED******REMOVED***.",
			usedIgnoredVar:
				"'{{varName***REMOVED******REMOVED***' is marked as ignored but is used{{additional***REMOVED******REMOVED***.",
			removeVar: "Remove unused variable '{{varName***REMOVED******REMOVED***'.",
		***REMOVED***,
	***REMOVED***,

	create(context) {
		const sourceCode = context.sourceCode;

		const REST_PROPERTY_TYPE =
			/^(?:RestElement|(?:Experimental)?RestProperty)$/u;

		const config = {
			vars: "all",
			args: "after-used",
			ignoreRestSiblings: false,
			caughtErrors: "all",
			ignoreClassWithStaticInitBlock: false,
			reportUsedIgnorePattern: false,
		***REMOVED***;

		const firstOption = context.options[0];

		if (firstOption) {
			if (typeof firstOption === "string") {
				config.vars = firstOption;
			***REMOVED*** else {
				config.vars = firstOption.vars || config.vars;
				config.args = firstOption.args || config.args;
				config.ignoreRestSiblings =
					firstOption.ignoreRestSiblings || config.ignoreRestSiblings;
				config.caughtErrors =
					firstOption.caughtErrors || config.caughtErrors;
				config.ignoreClassWithStaticInitBlock =
					firstOption.ignoreClassWithStaticInitBlock ||
					config.ignoreClassWithStaticInitBlock;
				config.reportUsedIgnorePattern =
					firstOption.reportUsedIgnorePattern ||
					config.reportUsedIgnorePattern;

				if (firstOption.varsIgnorePattern) {
					config.varsIgnorePattern = new RegExp(
						firstOption.varsIgnorePattern,
						"u",
					);
				***REMOVED***

				if (firstOption.argsIgnorePattern) {
					config.argsIgnorePattern = new RegExp(
						firstOption.argsIgnorePattern,
						"u",
					);
				***REMOVED***

				if (firstOption.caughtErrorsIgnorePattern) {
					config.caughtErrorsIgnorePattern = new RegExp(
						firstOption.caughtErrorsIgnorePattern,
						"u",
					);
				***REMOVED***

				if (firstOption.destructuredArrayIgnorePattern) {
					config.destructuredArrayIgnorePattern = new RegExp(
						firstOption.destructuredArrayIgnorePattern,
						"u",
					);
				***REMOVED***
			***REMOVED***
		***REMOVED***

		/**
		 * Determines what variable type a def is.
		 * @param  {Object***REMOVED*** def the declaration to check
		 * @returns {VariableType***REMOVED*** a simple name for the types of variables that this rule supports
		 */
		function defToVariableType(def) {
			/*
			 * This `destructuredArrayIgnorePattern` error report works differently from the catch
			 * clause and parameter error reports. _Both_ the `varsIgnorePattern` and the
			 * `destructuredArrayIgnorePattern` will be checked for array destructuring. However,
			 * for the purposes of the report, the currently defined behavior is to only inform the
			 * user of the `destructuredArrayIgnorePattern` if it's present (regardless of the fact
			 * that the `varsIgnorePattern` would also apply). If it's not present, the user will be
			 * informed of the `varsIgnorePattern`, assuming that's present.
			 */
			if (
				config.destructuredArrayIgnorePattern &&
				def.name.parent.type === "ArrayPattern"
			) {
				return "array-destructure";
			***REMOVED***

			switch (def.type) {
				case "CatchClause":
					return "catch-clause";
				case "Parameter":
					return "parameter";

				default:
					return "variable";
			***REMOVED***
		***REMOVED***

		/**
		 * Gets a given variable's description and configured ignore pattern
		 * based on the provided variableType
		 * @param {VariableType***REMOVED*** variableType a simple name for the types of variables that this rule supports
		 * @throws {Error***REMOVED*** (Unreachable)
		 * @returns {[string | undefined, string | undefined]***REMOVED*** the given variable's description and
		 * ignore pattern
		 */
		function getVariableDescription(variableType) {
			let pattern;
			let variableDescription;

			switch (variableType) {
				case "array-destructure":
					pattern = config.destructuredArrayIgnorePattern;
					variableDescription = "elements of array destructuring";
					break;

				case "catch-clause":
					pattern = config.caughtErrorsIgnorePattern;
					variableDescription = "caught errors";
					break;

				case "parameter":
					pattern = config.argsIgnorePattern;
					variableDescription = "args";
					break;

				case "variable":
					pattern = config.varsIgnorePattern;
					variableDescription = "vars";
					break;

				default:
					throw new Error(
						`Unexpected variable type: ${variableType***REMOVED***`,
					);
			***REMOVED***

			if (pattern) {
				pattern = pattern.toString();
			***REMOVED***

			return [variableDescription, pattern];
		***REMOVED***

		/**
		 * Generates the message data about the variable being defined and unused,
		 * including the ignore pattern if configured.
		 * @param {Variable***REMOVED*** unusedVar eslint-scope variable object.
		 * @returns {UnusedVarMessageData***REMOVED*** The message data to be used with this unused variable.
		 */
		function getDefinedMessageData(unusedVar) {
			const def = unusedVar.defs && unusedVar.defs[0];
			let additionalMessageData = "";

			if (def) {
				const [variableDescription, pattern] = getVariableDescription(
					defToVariableType(def),
				);

				if (pattern && variableDescription) {
					additionalMessageData = `. Allowed unused ${variableDescription***REMOVED*** must match ${pattern***REMOVED***`;
				***REMOVED***
			***REMOVED***

			return {
				varName: unusedVar.name,
				action: "defined",
				additional: additionalMessageData,
			***REMOVED***;
		***REMOVED***

		/**
		 * Generate the warning message about the variable being
		 * assigned and unused, including the ignore pattern if configured.
		 * @param {Variable***REMOVED*** unusedVar eslint-scope variable object.
		 * @returns {UnusedVarMessageData***REMOVED*** The message data to be used with this unused variable.
		 */
		function getAssignedMessageData(unusedVar) {
			const def = unusedVar.defs && unusedVar.defs[0];
			let additionalMessageData = "";

			if (def) {
				const [variableDescription, pattern] = getVariableDescription(
					defToVariableType(def),
				);

				if (pattern && variableDescription) {
					additionalMessageData = `. Allowed unused ${variableDescription***REMOVED*** must match ${pattern***REMOVED***`;
				***REMOVED***
			***REMOVED***

			return {
				varName: unusedVar.name,
				action: "assigned a value",
				additional: additionalMessageData,
			***REMOVED***;
		***REMOVED***

		/**
		 * Generate the warning message about a variable being used even though
		 * it is marked as being ignored.
		 * @param {Variable***REMOVED*** variable eslint-scope variable object
		 * @param {VariableType***REMOVED*** variableType a simple name for the types of variables that this rule supports
		 * @returns {UsedIgnoredVarMessageData***REMOVED*** The message data to be used with
		 * this used ignored variable.
		 */
		function getUsedIgnoredMessageData(variable, variableType) {
			const [variableDescription, pattern] =
				getVariableDescription(variableType);

			let additionalMessageData = "";

			if (pattern && variableDescription) {
				additionalMessageData = `. Used ${variableDescription***REMOVED*** must not match ${pattern***REMOVED***`;
			***REMOVED***

			return {
				varName: variable.name,
				additional: additionalMessageData,
			***REMOVED***;
		***REMOVED***

		//--------------------------------------------------------------------------
		// Helpers
		//--------------------------------------------------------------------------

		const STATEMENT_TYPE = /(?:Statement|Declaration)$/u;

		/**
		 * Determines if a given variable is being exported from a module.
		 * @param {Variable***REMOVED*** variable eslint-scope variable object.
		 * @returns {boolean***REMOVED*** True if the variable is exported, false if not.
		 * @private
		 */
		function isExported(variable) {
			const definition = variable.defs[0];

			if (definition) {
				let node = definition.node;

				if (node.type === "VariableDeclarator") {
					node = node.parent;
				***REMOVED*** else if (definition.type === "Parameter") {
					return false;
				***REMOVED***

				return node.parent.type.indexOf("Export") === 0;
			***REMOVED***
			return false;
		***REMOVED***

		/**
		 * Checks whether a node is a sibling of the rest property or not.
		 * @param {ASTNode***REMOVED*** node a node to check
		 * @returns {boolean***REMOVED*** True if the node is a sibling of the rest property, otherwise false.
		 */
		function hasRestSibling(node) {
			return (
				node.type === "Property" &&
				node.parent.type === "ObjectPattern" &&
				REST_PROPERTY_TYPE.test(node.parent.properties.at(-1).type)
			);
		***REMOVED***

		/**
		 * Determines if a variable has a sibling rest property
		 * @param {Variable***REMOVED*** variable eslint-scope variable object.
		 * @returns {boolean***REMOVED*** True if the variable has a sibling rest property, false if not.
		 * @private
		 */
		function hasRestSpreadSibling(variable) {
			if (config.ignoreRestSiblings) {
				const hasRestSiblingDefinition = variable.defs.some(def =>
					hasRestSibling(def.name.parent),
				);
				const hasRestSiblingReference = variable.references.some(ref =>
					hasRestSibling(ref.identifier.parent),
				);

				return hasRestSiblingDefinition || hasRestSiblingReference;
			***REMOVED***

			return false;
		***REMOVED***

		/**
		 * Determines if a reference is a read operation.
		 * @param {Reference***REMOVED*** ref An eslint-scope Reference
		 * @returns {boolean***REMOVED*** whether the given reference represents a read operation
		 * @private
		 */
		function isReadRef(ref) {
			return ref.isRead();
		***REMOVED***

		/**
		 * Determine if an identifier is referencing an enclosing function name.
		 * @param {Reference***REMOVED*** ref The reference to check.
		 * @param {ASTNode[]***REMOVED*** nodes The candidate function nodes.
		 * @returns {boolean***REMOVED*** True if it's a self-reference, false if not.
		 * @private
		 */
		function isSelfReference(ref, nodes) {
			let scope = ref.from;

			while (scope) {
				if (nodes.includes(scope.block)) {
					return true;
				***REMOVED***

				scope = scope.upper;
			***REMOVED***

			return false;
		***REMOVED***

		/**
		 * Gets a list of function definitions for a specified variable.
		 * @param {Variable***REMOVED*** variable eslint-scope variable object.
		 * @returns {ASTNode[]***REMOVED*** Function nodes.
		 * @private
		 */
		function getFunctionDefinitions(variable) {
			const functionDefinitions = [];

			variable.defs.forEach(def => {
				const { type, node ***REMOVED*** = def;

				// FunctionDeclarations
				if (type === "FunctionName") {
					functionDefinitions.push(node);
				***REMOVED***

				// FunctionExpressions
				if (
					type === "Variable" &&
					node.init &&
					(node.init.type === "FunctionExpression" ||
						node.init.type === "ArrowFunctionExpression")
				) {
					functionDefinitions.push(node.init);
				***REMOVED***
			***REMOVED***);
			return functionDefinitions;
		***REMOVED***

		/**
		 * Checks the position of given nodes.
		 * @param {ASTNode***REMOVED*** inner A node which is expected as inside.
		 * @param {ASTNode***REMOVED*** outer A node which is expected as outside.
		 * @returns {boolean***REMOVED*** `true` if the `inner` node exists in the `outer` node.
		 * @private
		 */
		function isInside(inner, outer) {
			return (
				inner.range[0] >= outer.range[0] &&
				inner.range[1] <= outer.range[1]
			);
		***REMOVED***

		/**
		 * Checks whether a given node is unused expression or not.
		 * @param {ASTNode***REMOVED*** node The node itself
		 * @returns {boolean***REMOVED*** The node is an unused expression.
		 * @private
		 */
		function isUnusedExpression(node) {
			const parent = node.parent;

			if (parent.type === "ExpressionStatement") {
				return true;
			***REMOVED***

			if (parent.type === "SequenceExpression") {
				const isLastExpression = parent.expressions.at(-1) === node;

				if (!isLastExpression) {
					return true;
				***REMOVED***
				return isUnusedExpression(parent);
			***REMOVED***

			return false;
		***REMOVED***

		/**
		 * If a given reference is left-hand side of an assignment, this gets
		 * the right-hand side node of the assignment.
		 *
		 * In the following cases, this returns null.
		 *
		 * - The reference is not the LHS of an assignment expression.
		 * - The reference is inside of a loop.
		 * - The reference is inside of a function scope which is different from
		 *   the declaration.
		 * @param {eslint-scope.Reference***REMOVED*** ref A reference to check.
		 * @param {ASTNode***REMOVED*** prevRhsNode The previous RHS node.
		 * @returns {ASTNode|null***REMOVED*** The RHS node or null.
		 * @private
		 */
		function getRhsNode(ref, prevRhsNode) {
			const id = ref.identifier;
			const parent = id.parent;
			const refScope = ref.from.variableScope;
			const varScope = ref.resolved.scope.variableScope;
			const canBeUsedLater =
				refScope !== varScope || astUtils.isInLoop(id);

			/*
			 * Inherits the previous node if this reference is in the node.
			 * This is for `a = a + a`-like code.
			 */
			if (prevRhsNode && isInside(id, prevRhsNode)) {
				return prevRhsNode;
			***REMOVED***

			if (
				parent.type === "AssignmentExpression" &&
				isUnusedExpression(parent) &&
				id === parent.left &&
				!canBeUsedLater
			) {
				return parent.right;
			***REMOVED***
			return null;
		***REMOVED***

		/**
		 * Checks whether a given function node is stored to somewhere or not.
		 * If the function node is stored, the function can be used later.
		 * @param {ASTNode***REMOVED*** funcNode A function node to check.
		 * @param {ASTNode***REMOVED*** rhsNode The RHS node of the previous assignment.
		 * @returns {boolean***REMOVED*** `true` if under the following conditions:
		 *      - the funcNode is assigned to a variable.
		 *      - the funcNode is bound as an argument of a function call.
		 *      - the function is bound to a property and the object satisfies above conditions.
		 * @private
		 */
		function isStorableFunction(funcNode, rhsNode) {
			let node = funcNode;
			let parent = funcNode.parent;

			while (parent && isInside(parent, rhsNode)) {
				switch (parent.type) {
					case "SequenceExpression":
						if (parent.expressions.at(-1) !== node) {
							return false;
						***REMOVED***
						break;

					case "CallExpression":
					case "NewExpression":
						return parent.callee !== node;

					case "AssignmentExpression":
					case "TaggedTemplateExpression":
					case "YieldExpression":
						return true;

					default:
						if (STATEMENT_TYPE.test(parent.type)) {
							/*
							 * If it encountered statements, this is a complex pattern.
							 * Since analyzing complex patterns is hard, this returns `true` to avoid false positive.
							 */
							return true;
						***REMOVED***
				***REMOVED***

				node = parent;
				parent = parent.parent;
			***REMOVED***

			return false;
		***REMOVED***

		/**
		 * Checks whether a given Identifier node exists inside of a function node which can be used later.
		 *
		 * "can be used later" means:
		 * - the function is assigned to a variable.
		 * - the function is bound to a property and the object can be used later.
		 * - the function is bound as an argument of a function call.
		 *
		 * If a reference exists in a function which can be used later, the reference is read when the function is called.
		 * @param {ASTNode***REMOVED*** id An Identifier node to check.
		 * @param {ASTNode***REMOVED*** rhsNode The RHS node of the previous assignment.
		 * @returns {boolean***REMOVED*** `true` if the `id` node exists inside of a function node which can be used later.
		 * @private
		 */
		function isInsideOfStorableFunction(id, rhsNode) {
			const funcNode = astUtils.getUpperFunction(id);

			return (
				funcNode &&
				isInside(funcNode, rhsNode) &&
				isStorableFunction(funcNode, rhsNode)
			);
		***REMOVED***

		/**
		 * Checks whether a given reference is a read to update itself or not.
		 * @param {eslint-scope.Reference***REMOVED*** ref A reference to check.
		 * @param {ASTNode***REMOVED*** rhsNode The RHS node of the previous assignment.
		 * @returns {boolean***REMOVED*** The reference is a read to update itself.
		 * @private
		 */
		function isReadForItself(ref, rhsNode) {
			const id = ref.identifier;
			const parent = id.parent;

			return (
				ref.isRead() &&
				// self update. e.g. `a += 1`, `a++`
				((parent.type === "AssignmentExpression" &&
					parent.left === id &&
					isUnusedExpression(parent) &&
					!astUtils.isLogicalAssignmentOperator(parent.operator)) ||
					(parent.type === "UpdateExpression" &&
						isUnusedExpression(parent)) ||
					// in RHS of an assignment for itself. e.g. `a = a + 1`
					(rhsNode &&
						isInside(id, rhsNode) &&
						!isInsideOfStorableFunction(id, rhsNode)))
			);
		***REMOVED***

		/**
		 * Determine if an identifier is used either in for-in or for-of loops.
		 * @param {Reference***REMOVED*** ref The reference to check.
		 * @returns {boolean***REMOVED*** whether reference is used in the for-in loops
		 * @private
		 */
		function isForInOfRef(ref) {
			let target = ref.identifier.parent;

			// "for (var ...) { return; ***REMOVED***"
			if (target.type === "VariableDeclarator") {
				target = target.parent.parent;
			***REMOVED***

			if (
				target.type !== "ForInStatement" &&
				target.type !== "ForOfStatement"
			) {
				return false;
			***REMOVED***

			// "for (...) { return; ***REMOVED***"
			if (target.body.type === "BlockStatement") {
				target = target.body.body[0];

				// "for (...) return;"
			***REMOVED*** else {
				target = target.body;
			***REMOVED***

			// For empty loop body
			if (!target) {
				return false;
			***REMOVED***

			return target.type === "ReturnStatement";
		***REMOVED***

		/**
		 * Determines if the variable is used.
		 * @param {Variable***REMOVED*** variable The variable to check.
		 * @returns {boolean***REMOVED*** True if the variable is used
		 * @private
		 */
		function isUsedVariable(variable) {
			if (variable.eslintUsed) {
				return true;
			***REMOVED***

			const functionNodes = getFunctionDefinitions(variable);
			const isFunctionDefinition = functionNodes.length > 0;

			let rhsNode = null;

			return variable.references.some(ref => {
				if (isForInOfRef(ref)) {
					return true;
				***REMOVED***

				const forItself = isReadForItself(ref, rhsNode);

				rhsNode = getRhsNode(ref, rhsNode);

				return (
					isReadRef(ref) &&
					!forItself &&
					!(
						isFunctionDefinition &&
						isSelfReference(ref, functionNodes)
					)
				);
			***REMOVED***);
		***REMOVED***

		/**
		 * Checks whether the given variable is after the last used parameter.
		 * @param {eslint-scope.Variable***REMOVED*** variable The variable to check.
		 * @returns {boolean***REMOVED*** `true` if the variable is defined after the last
		 * used parameter.
		 */
		function isAfterLastUsedArg(variable) {
			const def = variable.defs[0];
			const params = sourceCode.getDeclaredVariables(def.node);
			const posteriorParams = params.slice(params.indexOf(variable) + 1);

			// If any used parameters occur after this parameter, do not report.
			return !posteriorParams.some(
				v => v.references.length > 0 || v.eslintUsed,
			);
		***REMOVED***

		/**
		 * Gets an array of variables without read references.
		 * @param {Scope***REMOVED*** scope an eslint-scope Scope object.
		 * @param {Variable[]***REMOVED*** unusedVars an array that saving result.
		 * @returns {Variable[]***REMOVED*** unused variables of the scope and descendant scopes.
		 * @private
		 */
		function collectUnusedVariables(scope, unusedVars) {
			const variables = scope.variables;
			const childScopes = scope.childScopes;
			let i, l;

			if (scope.type !== "global" || config.vars === "all") {
				for (i = 0, l = variables.length; i < l; ++i) {
					const variable = variables[i];

					// skip a variable of class itself name in the class scope
					if (
						scope.type === "class" &&
						scope.block.id === variable.identifiers[0]
					) {
						continue;
					***REMOVED***

					// skip function expression names
					if (scope.functionExpressionScope) {
						continue;
					***REMOVED***

					// skip variables marked with markVariableAsUsed()
					if (
						!config.reportUsedIgnorePattern &&
						variable.eslintUsed
					) {
						continue;
					***REMOVED***

					// skip implicit "arguments" variable
					if (
						scope.type === "function" &&
						variable.name === "arguments" &&
						variable.identifiers.length === 0
					) {
						continue;
					***REMOVED***

					// explicit global variables don't have definitions.
					const def = variable.defs[0];

					if (def) {
						const type = def.type;
						const refUsedInArrayPatterns = variable.references.some(
							ref =>
								ref.identifier.parent.type === "ArrayPattern",
						);

						// skip elements of array destructuring patterns
						if (
							(def.name.parent.type === "ArrayPattern" ||
								refUsedInArrayPatterns) &&
							config.destructuredArrayIgnorePattern &&
							config.destructuredArrayIgnorePattern.test(
								def.name.name,
							)
						) {
							if (
								config.reportUsedIgnorePattern &&
								isUsedVariable(variable)
							) {
								context.report({
									node: def.name,
									messageId: "usedIgnoredVar",
									data: getUsedIgnoredMessageData(
										variable,
										"array-destructure",
									),
								***REMOVED***);
							***REMOVED***

							continue;
						***REMOVED***

						if (type === "ClassName") {
							const hasStaticBlock = def.node.body.body.some(
								node => node.type === "StaticBlock",
							);

							if (
								config.ignoreClassWithStaticInitBlock &&
								hasStaticBlock
							) {
								continue;
							***REMOVED***
						***REMOVED***

						// skip catch variables
						if (type === "CatchClause") {
							if (config.caughtErrors === "none") {
								continue;
							***REMOVED***

							// skip ignored parameters
							if (
								config.caughtErrorsIgnorePattern &&
								config.caughtErrorsIgnorePattern.test(
									def.name.name,
								)
							) {
								if (
									config.reportUsedIgnorePattern &&
									isUsedVariable(variable)
								) {
									context.report({
										node: def.name,
										messageId: "usedIgnoredVar",
										data: getUsedIgnoredMessageData(
											variable,
											"catch-clause",
										),
									***REMOVED***);
								***REMOVED***

								continue;
							***REMOVED***
						***REMOVED*** else if (type === "Parameter") {
							// skip any setter argument
							if (
								(def.node.parent.type === "Property" ||
									def.node.parent.type ===
										"MethodDefinition") &&
								def.node.parent.kind === "set"
							) {
								continue;
							***REMOVED***

							// if "args" option is "none", skip any parameter
							if (config.args === "none") {
								continue;
							***REMOVED***

							// skip ignored parameters
							if (
								config.argsIgnorePattern &&
								config.argsIgnorePattern.test(def.name.name)
							) {
								if (
									config.reportUsedIgnorePattern &&
									isUsedVariable(variable)
								) {
									context.report({
										node: def.name,
										messageId: "usedIgnoredVar",
										data: getUsedIgnoredMessageData(
											variable,
											"parameter",
										),
									***REMOVED***);
								***REMOVED***

								continue;
							***REMOVED***

							// if "args" option is "after-used", skip used variables
							if (
								config.args === "after-used" &&
								astUtils.isFunction(def.name.parent) &&
								!isAfterLastUsedArg(variable)
							) {
								continue;
							***REMOVED***
						***REMOVED*** else {
							// skip ignored variables
							if (
								config.varsIgnorePattern &&
								config.varsIgnorePattern.test(def.name.name)
							) {
								if (
									config.reportUsedIgnorePattern &&
									isUsedVariable(variable)
								) {
									context.report({
										node: def.name,
										messageId: "usedIgnoredVar",
										data: getUsedIgnoredMessageData(
											variable,
											"variable",
										),
									***REMOVED***);
								***REMOVED***

								continue;
							***REMOVED***
						***REMOVED***
					***REMOVED***

					if (
						!isUsedVariable(variable) &&
						!isExported(variable) &&
						!hasRestSpreadSibling(variable)
					) {
						unusedVars.push(variable);
					***REMOVED***
				***REMOVED***
			***REMOVED***

			for (i = 0, l = childScopes.length; i < l; ++i) {
				collectUnusedVariables(childScopes[i], unusedVars);
			***REMOVED***

			return unusedVars;
		***REMOVED***

		/**
		 * fixes unused variables
		 * @param {Object***REMOVED*** fixer fixer object
		 * @param {Object***REMOVED*** unusedVar unused variable to fix
		 * @returns {Object***REMOVED*** fixer object
		 */
		function handleFixes(fixer, unusedVar) {
			const id = unusedVar.identifiers[0];
			const parent = id.parent;
			const parentType = parent.type;
			const tokenBefore = sourceCode.getTokenBefore(id);
			const tokenAfter = sourceCode.getTokenAfter(id);
			const isFunction = astUtils.isFunction;
			const isLoop = astUtils.isLoop;
			const allWriteReferences = unusedVar.references.filter(ref =>
				ref.isWrite(),
			);

			/**
			 * get range from token before of a given node
			 * @param {ASTNode***REMOVED*** node node of identifier
			 * @param {number***REMOVED*** skips number of token to skip
			 * @returns {number***REMOVED*** start range of token before the identifier
			 */
			function getPreviousTokenStart(node, skips) {
				return sourceCode.getTokenBefore(node, skips).range[0];
			***REMOVED***

			/**
			 * get range to token after of a given node
			 * @param {ASTNode***REMOVED*** node node of identifier
			 * @param {number***REMOVED*** skips number of token to skip
			 * @returns {number***REMOVED*** end range of token after the identifier
			 */
			function getNextTokenEnd(node, skips) {
				return sourceCode.getTokenAfter(node, skips).range[1];
			***REMOVED***

			/**
			 * get the value of token before of a given node
			 * @param {ASTNode***REMOVED*** node node of identifier
			 * @returns {string***REMOVED*** value of token before the identifier
			 */
			function getTokenBeforeValue(node) {
				return sourceCode.getTokenBefore(node).value;
			***REMOVED***

			/**
			 * get the value of token after of a given node
			 * @param {ASTNode***REMOVED*** node node of identifier
			 * @returns {string***REMOVED*** value of token after the identifier
			 */
			function getTokenAfterValue(node) {
				return sourceCode.getTokenAfter(node).value;
			***REMOVED***

			/**
			 * Check if an array has a single element with null as other element.
			 * @param {ASTNode***REMOVED*** node ArrayPattern node
			 * @returns {boolean***REMOVED*** true if array has single element with other null elements
			 */
			function hasSingleElement(node) {
				return node.elements.filter(e => e !== null).length === 1;
			***REMOVED***

			/**
			 * check whether import specifier has an import of particular type
			 * @param {ASTNode***REMOVED*** node ImportDeclaration node
			 * @param {string***REMOVED*** type type of import to check
			 * @returns {boolean***REMOVED*** true if import specifier has import of specified type
			 */
			function hasImportOfCertainType(node, type) {
				return node.specifiers.some(e => e.type === type);
			***REMOVED***

			/**
			 * Check whether declaration is safe to remove or not
			 * @param {ASTNode***REMOVED*** nextToken next token of unused variable
			 * @param {ASTNode***REMOVED*** prevToken previous token of unused variable
			 * @returns {boolean***REMOVED*** true if declaration is not safe to remove
			 */
			function isDeclarationNotSafeToRemove(nextToken, prevToken) {
				return (
					nextToken.type === "String" ||
					(prevToken &&
						!astUtils.isSemicolonToken(prevToken) &&
						!astUtils.isOpeningBraceToken(prevToken))
				);
			***REMOVED***

			/**
			 * give fixes for unused variables in function parameters
			 * @param {ASTNode***REMOVED*** node node to check
			 * @returns {Object***REMOVED*** fixer object
			 */
			function fixFunctionParameters(node) {
				const parentNode = node.parent;

				if (isFunction(parentNode)) {
					// remove unused function parameter if there is only a single parameter
					if (parentNode.params.length === 1) {
						return fixer.removeRange(node.range);
					***REMOVED***

					// remove first unused function parameter when there are multiple parameters
					if (
						getTokenBeforeValue(node) === "(" &&
						getTokenAfterValue(node) === ","
					) {
						return fixer.removeRange([
							node.range[0],
							getNextTokenEnd(node),
						]);
					***REMOVED***

					// remove unused function parameters except first one when there are multiple parameters
					return fixer.removeRange([
						getPreviousTokenStart(node),
						node.range[1],
					]);
				***REMOVED***

				return null;
			***REMOVED***

			/**
			 * fix unused variable declarations and function parameters
			 * @param {ASTNode***REMOVED*** node parent node to identifier
			 * @returns {Object***REMOVED*** fixer object
			 */
			function fixVariables(node) {
				const parentNode = node.parent;

				// remove unused declared variables such as var a = b; or var a = b, c;
				if (parentNode.type === "VariableDeclarator") {
					// skip variable in for (const [ foo ] of bar);
					if (isLoop(parentNode.parent.parent)) {
						return null;
					***REMOVED***

					/*
					 * remove unused declared variable with single declaration such as 'var a = b;'
					 * remove complete declaration when there is an unused variable in 'const { a ***REMOVED*** = foo;', same for arrays.
					 */
					if (parentNode.parent.declarations.length === 1) {
						// if next token is a string it could become a directive if node is removed -> no suggestion.
						const nextToken = sourceCode.getTokenAfter(
							parentNode.parent,
						);

						// if previous token exists and is not ";" or "{" not sure about ASI rules -> no suggestion.
						const prevToken = sourceCode.getTokenBefore(
							parentNode.parent,
						);

						if (
							nextToken &&
							isDeclarationNotSafeToRemove(nextToken, prevToken)
						) {
							return null;
						***REMOVED***

						return fixer.removeRange(parentNode.parent.range);
					***REMOVED***

					/*
					 * remove unused declared variable with multiple declaration except first one such as 'var a = b, c = d;'
					 * fix 'let bar = "hello", { a ***REMOVED*** = foo;' to 'let bar = "hello";' if 'a' is unused, same for arrays.
					 */
					if (getTokenBeforeValue(parentNode) === ",") {
						return fixer.removeRange([
							getPreviousTokenStart(parentNode),
							parentNode.range[1],
						]);
					***REMOVED***

					/*
					 * remove first unused declared variable when there are multiple declarations
					 * fix 'let { a ***REMOVED*** = foo, bar = "hello";' to 'let  bar = "hello";' if 'a' is unused, same for arrays.
					 */
					return fixer.removeRange([
						parentNode.range[0],
						getNextTokenEnd(parentNode),
					]);
				***REMOVED***

				// fixes [{a: {k***REMOVED******REMOVED***], [{a: [k]***REMOVED***]
				if (getTokenBeforeValue(node) === ":") {
					if (parentNode.parent.type === "ObjectPattern") {
						// eslint-disable-next-line no-use-before-define -- due to interdependency of functions
						return fixObjectWithValueSeparator(node);
					***REMOVED***
				***REMOVED***

				// fix unused function parameters
				return fixFunctionParameters(node);
			***REMOVED***

			/**
			 * fix nested object like { a: { b ***REMOVED*** ***REMOVED***
			 * @param {ASTNode***REMOVED*** node parent node to check
			 * @returns {Object***REMOVED*** fixer object
			 */
			function fixNestedObjectVariable(node) {
				const parentNode = node.parent;

				// fix for { a: { b: { c: { d ***REMOVED*** ***REMOVED*** ***REMOVED*** ***REMOVED***
				if (
					parentNode.parent.parent.parent.type === "ObjectPattern" &&
					parentNode.parent.properties.length === 1
				) {
					return fixNestedObjectVariable(parentNode.parent);
				***REMOVED***

				// fix for { a: { b ***REMOVED*** ***REMOVED***
				if (parentNode.parent.type === "ObjectPattern") {
					// fix for unused variables in dectructured object with single property in variable decalartion and function parameter
					if (parentNode.parent.properties.length === 1) {
						return fixVariables(parentNode.parent);
					***REMOVED***

					// fix for first unused property when there are multiple properties such as '{ a: { b ***REMOVED***, c ***REMOVED***'
					if (getTokenBeforeValue(parentNode) === "{") {
						return fixer.removeRange([
							parentNode.range[0],
							getNextTokenEnd(parentNode),
						]);
					***REMOVED***

					// fix for unused property except first one when there are multiple properties such as '{ k, a: { b ***REMOVED*** ***REMOVED***'
					return fixer.removeRange([
						getPreviousTokenStart(parentNode),
						parentNode.range[1],
					]);
				***REMOVED***

				return null;
			***REMOVED***

			/**
			 * fix unused variables in array and nested array
			 * @param {ASTNode***REMOVED*** node parent node to check
			 * @returns {Object***REMOVED*** fixer object
			 */
			function fixNestedArrayVariable(node) {
				const parentNode = node.parent;

				// fix for nested arrays [[ a ]]
				if (
					parentNode.parent.type === "ArrayPattern" &&
					hasSingleElement(parentNode)
				) {
					return fixNestedArrayVariable(parentNode);
				***REMOVED***

				if (hasSingleElement(parentNode)) {
					// fixes { a: [{ b ***REMOVED***] ***REMOVED*** or { a: [[ b ]] ***REMOVED***
					if (getTokenBeforeValue(parentNode) === ":") {
						return fixVariables(parentNode);
					***REMOVED***

					// fixes [a, ...[[ b ]]] or [a, ...[{ b ***REMOVED***]]
					if (parentNode.parent.type === "RestElement") {
						// eslint-disable-next-line no-use-before-define -- due to interdependency of functions
						return fixRestInPattern(parentNode.parent);
					***REMOVED***

					// fix unused variables in destructured array in variable declaration or function parameter
					return fixVariables(parentNode);
				***REMOVED***

				// remove last unused array element
				if (
					getTokenBeforeValue(node) === "," &&
					getTokenAfterValue(node) === "]"
				) {
					return fixer.removeRange([
						getPreviousTokenStart(node),
						node.range[1],
					]);
				***REMOVED***

				// remove unused array element
				return fixer.removeRange(node.range);
			***REMOVED***

			/**
			 * fix cases like {a: {k***REMOVED******REMOVED*** or {a: [k]***REMOVED***
			 * @param {ASTNode***REMOVED*** node parent node to check
			 * @returns {Object***REMOVED*** fixer object
			 */
			function fixObjectWithValueSeparator(node) {
				const parentNode = node.parent.parent;

				// fix cases like [{a : { b ***REMOVED******REMOVED***] or [{a : [ b ]***REMOVED***]
				if (
					parentNode.parent.type === "ArrayPattern" &&
					parentNode.properties.length === 1
				) {
					return fixNestedArrayVariable(parentNode);
				***REMOVED***

				// fix cases like {a: {k***REMOVED******REMOVED*** or {a: [k]***REMOVED***
				return fixNestedObjectVariable(node);
			***REMOVED***

			/**
			 * fix ...[[a]] or ...[{a***REMOVED***] like patterns
			 * @param {ASTNode***REMOVED*** node parent node to check
			 * @returns {Object***REMOVED*** fixer object
			 */
			function fixRestInPattern(node) {
				const parentNode = node.parent;

				// fix ...[[a]] or ...[{a***REMOVED***] in function parameters
				if (isFunction(parentNode)) {
					if (parentNode.params.length === 1) {
						return fixer.removeRange(node.range);
					***REMOVED***

					return fixer.removeRange([
						getPreviousTokenStart(node),
						node.range[1],
					]);
				***REMOVED***

				// fix rest in nested array pattern like [[a, ...[b]]]
				if (parentNode.type === "ArrayPattern") {
					// fix [[...[b]]]
					if (hasSingleElement(parentNode)) {
						if (parentNode.parent.type === "ArrayPattern") {
							return fixNestedArrayVariable(parentNode);
						***REMOVED***

						// fix 'const [...[b]] = foo; and function foo([...[b]]) {***REMOVED***
						return fixVariables(parentNode);
					***REMOVED***

					// fix [[a, ...[b]]]
					return fixer.removeRange([
						getPreviousTokenStart(node),
						node.range[1],
					]);
				***REMOVED***

				return null;
			***REMOVED***

			// skip fix when variable has references that would be left behind
			if (
				allWriteReferences.some(
					ref => ref.identifier.range[0] !== id.range[0],
				)
			) {
				return null;
			***REMOVED***

			// remove declared variables such as var a; or var a, b;
			if (parentType === "VariableDeclarator") {
				if (parent.parent.declarations.length === 1) {
					// prevent fix of variable in forOf and forIn loops.
					if (
						isLoop(parent.parent.parent) &&
						parent.parent.parent.body !== parent.parent
					) {
						return null;
					***REMOVED***

					// removes only variable not semicolon in 'if (foo()) var bar;' or in 'loops' or in 'with' statement.
					if (
						parent.parent.parent.type === "IfStatement" ||
						isLoop(parent.parent.parent) ||
						(parent.parent.parent.type === "WithStatement" &&
							parent.parent.parent.body === parent.parent)
					) {
						return fixer.replaceText(parent.parent, ";");
					***REMOVED***

					// if next token is a string it could become a directive if node is removed -> no suggestion.
					const nextToken = sourceCode.getTokenAfter(parent.parent);

					// if previous token exists and is not ";" or "{" not sure about ASI rules -> no suggestion.
					const prevToken = sourceCode.getTokenBefore(parent.parent);

					if (
						nextToken &&
						isDeclarationNotSafeToRemove(nextToken, prevToken)
					) {
						return null;
					***REMOVED***

					// remove unused declared variable with single declaration like 'var a = b;'
					return fixer.removeRange(parent.parent.range);
				***REMOVED***

				// remove unused declared variable with multiple declaration except first one like 'var a = b, c = d;'
				if (tokenBefore.value === ",") {
					return fixer.removeRange([
						tokenBefore.range[0],
						parent.range[1],
					]);
				***REMOVED***

				// remove first unused declared variable when there are multiple declarations
				return fixer.removeRange([
					parent.range[0],
					getNextTokenEnd(parent),
				]);
			***REMOVED***

			// remove variables in object patterns
			if (parent.parent.type === "ObjectPattern") {
				if (parent.parent.properties.length === 1) {
					// fix [a, ...{b***REMOVED***]
					if (parent.parent.parent.type === "RestElement") {
						return fixRestInPattern(parent.parent.parent);
					***REMOVED***

					// fix [{ a ***REMOVED***]
					if (parent.parent.parent.type === "ArrayPattern") {
						return fixNestedArrayVariable(parent.parent);
					***REMOVED***

					/*
					 * var {a***REMOVED*** = foo;
					 * function a({a***REMOVED***) {***REMOVED***
					 * fix const { a: { b ***REMOVED*** ***REMOVED*** = foo;
					 */
					return fixVariables(parent.parent);
				***REMOVED***

				// fix const { a:b ***REMOVED*** = foo;
				if (tokenBefore.value === ":") {
					// remove first unused variable in const { a:b ***REMOVED*** = foo;
					if (
						getTokenBeforeValue(parent) === "{" &&
						getTokenAfterValue(parent) === ","
					) {
						return fixer.removeRange([
							parent.range[0],
							getNextTokenEnd(parent),
						]);
					***REMOVED***

					// remove unused variables in const { a: b, c: d ***REMOVED*** = foo; except first one
					return fixer.removeRange([
						getPreviousTokenStart(parent),
						id.range[1],
					]);
				***REMOVED***
			***REMOVED***

			// remove unused variables inside an array
			if (parentType === "ArrayPattern") {
				if (hasSingleElement(parent)) {
					// fix [a, ...[b]]
					if (parent.parent.type === "RestElement") {
						return fixRestInPattern(parent.parent);
					***REMOVED***

					// fix [ [a] ]
					if (parent.parent.type === "ArrayPattern") {
						return fixNestedArrayVariable(parent);
					***REMOVED***

					/*
					 * fix var [a] = foo;
					 * fix function foo([a]) {***REMOVED***
					 * fix const { a: [b] ***REMOVED*** = foo;
					 */
					return fixVariables(parent);
				***REMOVED***

				// if "a" is unused in [a, b ,c] fixes to [, b, c]
				if (tokenBefore.value === "," && tokenAfter.value === ",") {
					return fixer.removeRange(id.range);
				***REMOVED***
			***REMOVED***

			// remove unused rest elements
			if (parentType === "RestElement") {
				// fix [a, ...rest]
				if (parent.parent.type === "ArrayPattern") {
					if (hasSingleElement(parent.parent)) {
						// fix [[...rest]] when there is only rest element
						if (parent.parent.parent.type === "ArrayPattern") {
							return fixNestedArrayVariable(parent.parent);
						***REMOVED***

						// fix 'const [...rest] = foo;' and 'function foo([...rest]) {***REMOVED***'
						return fixVariables(parent.parent);
					***REMOVED***

					// fix [a, ...rest]
					return fixer.removeRange([
						getPreviousTokenStart(id, 1),
						id.range[1],
					]);
				***REMOVED***

				// fix { a, ...rest***REMOVED***
				if (parent.parent.type === "ObjectPattern") {
					// fix 'const {...rest***REMOVED*** = foo;' and 'function foo({...rest***REMOVED***) {***REMOVED***'
					if (parent.parent.properties.length === 1) {
						return fixVariables(parent.parent);
					***REMOVED***

					// fix { a, ...rest***REMOVED*** when there are multiple properties
					return fixer.removeRange([
						getPreviousTokenStart(id, 1),
						id.range[1],
					]);
				***REMOVED***

				// fix function foo(...rest) {***REMOVED***
				if (isFunction(parent.parent)) {
					// remove unused rest in function parameter if there is only single parameter
					if (parent.parent.params.length === 1) {
						return fixer.removeRange(parent.range);
					***REMOVED***

					// remove unused rest in function parameter if there multiple parameter
					return fixer.removeRange([
						getPreviousTokenStart(parent),
						parent.range[1],
					]);
				***REMOVED***
			***REMOVED***

			if (parentType === "AssignmentPattern") {
				// fix [a = aDefault]
				if (parent.parent.type === "ArrayPattern") {
					return fixNestedArrayVariable(parent);
				***REMOVED***

				// fix {a = aDefault***REMOVED***
				if (parent.parent.parent.type === "ObjectPattern") {
					if (parent.parent.parent.properties.length === 1) {
						// fixes [{a = aDefault***REMOVED***]
						if (
							parent.parent.parent.parent.type === "ArrayPattern"
						) {
							return fixNestedArrayVariable(parent.parent.parent);
						***REMOVED***

						// fix 'const {a = aDefault***REMOVED*** = foo;' and 'function foo({a = aDefault***REMOVED***) {***REMOVED***'
						return fixVariables(parent.parent.parent);
					***REMOVED***

					// fix unused 'a' in {a = aDefault***REMOVED*** if it is the first property
					if (
						getTokenBeforeValue(parent.parent) === "{" &&
						getTokenAfterValue(parent.parent) === ","
					) {
						return fixer.removeRange([
							parent.parent.range[0],
							getNextTokenEnd(parent.parent),
						]);
					***REMOVED***

					// fix unused 'b' in {a, b = aDefault***REMOVED*** if it is not the first property
					return fixer.removeRange([
						getPreviousTokenStart(parent.parent),
						parent.parent.range[1],
					]);
				***REMOVED***

				// fix unused assignment patterns in function parameters
				if (isFunction(parent.parent)) {
					return fixFunctionParameters(parent);
				***REMOVED***
			***REMOVED***

			// remove unused functions
			if (parentType === "FunctionDeclaration" && parent.id === id) {
				return fixer.removeRange(parent.range);
			***REMOVED***

			// remove unused default import
			if (parentType === "ImportDefaultSpecifier") {
				// remove unused default import when there are not other imports
				if (
					!hasImportOfCertainType(parent.parent, "ImportSpecifier") &&
					!hasImportOfCertainType(
						parent.parent,
						"ImportNamespaceSpecifier",
					)
				) {
					return fixer.removeRange([
						parent.range[0],
						parent.parent.source.range[0],
					]);
				***REMOVED***

				// remove unused default import when there are other imports also
				return fixer.removeRange([id.range[0], tokenAfter.range[1]]);
			***REMOVED***

			if (parentType === "ImportSpecifier") {
				// remove unused imports when there is a single import
				if (
					parent.parent.specifiers.filter(
						e => e.type === "ImportSpecifier",
					).length === 1
				) {
					// remove unused import when there is no default import
					if (
						!hasImportOfCertainType(
							parent.parent,
							"ImportDefaultSpecifier",
						)
					) {
						return fixer.removeRange(parent.parent.range);
					***REMOVED***

					// fixes "import foo from 'module';" to "import 'module';"
					return fixer.removeRange([
						getPreviousTokenStart(parent, 1),
						tokenAfter.range[1],
					]);
				***REMOVED***

				if (getTokenBeforeValue(parent) === "{") {
					return fixer.removeRange([
						parent.range[0],
						getNextTokenEnd(parent),
					]);
				***REMOVED***

				return fixer.removeRange([
					getPreviousTokenStart(parent),
					parent.range[1],
				]);
			***REMOVED***

			if (parentType === "ImportNamespaceSpecifier") {
				if (
					hasImportOfCertainType(
						parent.parent,
						"ImportDefaultSpecifier",
					)
				) {
					return fixer.removeRange([
						getPreviousTokenStart(parent),
						parent.range[1],
					]);
				***REMOVED***

				// fixes "import * as foo from 'module';" to "import 'module';"
				return fixer.removeRange([
					parent.range[0],
					parent.parent.source.range[0],
				]);
			***REMOVED***

			// skip error in catch(error) variable
			if (parentType === "CatchClause") {
				return null;
			***REMOVED***

			// remove unused declared classes
			if (parentType === "ClassDeclaration") {
				return fixer.removeRange(parent.range);
			***REMOVED***

			// remove unused variable that is in a sequence [a,b] fixes to [a]
			if (tokenBefore?.value === ",") {
				return fixer.removeRange([tokenBefore.range[0], id.range[1]]);
			***REMOVED***

			// remove unused variable that is in a sequence inside function arguments and object pattern
			if (tokenAfter.value === ",") {
				// fix function foo(a, b) {***REMOVED***
				if (tokenBefore.value === "(") {
					return fixer.removeRange([
						id.range[0],
						tokenAfter.range[1],
					]);
				***REMOVED***

				// fix const {a, b***REMOVED*** = foo;
				if (tokenBefore.value === "{") {
					return fixer.removeRange([
						id.range[0],
						tokenAfter.range[1],
					]);
				***REMOVED***
			***REMOVED***

			if (
				parentType === "ArrowFunctionExpression" &&
				parent.params.length === 1 &&
				tokenAfter?.value !== ")"
			) {
				return fixer.replaceText(id, "()");
			***REMOVED***

			return fixer.removeRange(id.range);
		***REMOVED***

		//--------------------------------------------------------------------------
		// Public
		//--------------------------------------------------------------------------

		return {
			"Program:exit"(programNode) {
				const unusedVars = collectUnusedVariables(
					sourceCode.getScope(programNode),
					[],
				);

				for (let i = 0, l = unusedVars.length; i < l; ++i) {
					const unusedVar = unusedVars[i];

					// Report the first declaration.
					if (unusedVar.defs.length > 0) {
						// report last write reference, https://github.com/eslint/eslint/issues/14324
						const writeReferences = unusedVar.references.filter(
							ref =>
								ref.isWrite() &&
								ref.from.variableScope ===
									unusedVar.scope.variableScope,
						);

						let referenceToReport;

						if (writeReferences.length > 0) {
							referenceToReport = writeReferences.at(-1);
						***REMOVED***

						context.report({
							node: referenceToReport
								? referenceToReport.identifier
								: unusedVar.identifiers[0],
							messageId: "unusedVar",
							data: unusedVar.references.some(ref =>
								ref.isWrite(),
							)
								? getAssignedMessageData(unusedVar)
								: getDefinedMessageData(unusedVar),
							suggest: [
								{
									messageId: "removeVar",
									data: {
										varName: unusedVar.name,
									***REMOVED***,
									fix(fixer) {
										return handleFixes(fixer, unusedVar);
									***REMOVED***,
								***REMOVED***,
							],
						***REMOVED***);

						// If there are no regular declaration, report the first `/*globals*/` comment directive.
					***REMOVED*** else if (unusedVar.eslintExplicitGlobalComments) {
						const directiveComment =
							unusedVar.eslintExplicitGlobalComments[0];

						context.report({
							node: programNode,
							loc: astUtils.getNameLocationInGlobalDirectiveComment(
								sourceCode,
								directiveComment,
								unusedVar.name,
							),
							messageId: "unusedVar",
							data: getDefinedMessageData(unusedVar),
						***REMOVED***);
					***REMOVED***
				***REMOVED***
			***REMOVED***,
		***REMOVED***;
	***REMOVED***,
***REMOVED***;
