// eslint/lib/rules/no-unsafe-finally.js
/**
 * @fileoverview Rule to flag unsafe statements in finally block
 * @author Onur Temizkan
 */

"use strict";

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

const SENTINEL_NODE_TYPE_RETURN_THROW = /^(?:Program|(?:Function|Class)(?:Declaration|Expression)|ArrowFunctionExpression)$/u;
const SENTINEL_NODE_TYPE_BREAK = /^(?:Program|(?:Function|Class)(?:Declaration|Expression)|ArrowFunctionExpression|DoWhileStatement|WhileStatement|ForOfStatement|ForInStatement|ForStatement|SwitchStatement)$/u;
const SENTINEL_NODE_TYPE_CONTINUE = /^(?:Program|(?:Function|Class)(?:Declaration|Expression)|ArrowFunctionExpression|DoWhileStatement|WhileStatement|ForOfStatement|ForInStatement|ForStatement)$/u;


//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "problem",

        docs: {
            description: "disallow control flow statements in `finally` blocks",
            category: "Possible Errors",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-unsafe-finally"
        ***REMOVED***,

        schema: [],

        messages: {
            unsafeUsage: "Unsafe usage of {{nodeType***REMOVED******REMOVED***."
        ***REMOVED***
    ***REMOVED***,
    create(context) {

        /**
         * Checks if the node is the finalizer of a TryStatement
         * @param {ASTNode***REMOVED*** node node to check.
         * @returns {boolean***REMOVED*** - true if the node is the finalizer of a TryStatement
         */
        function isFinallyBlock(node) {
            return node.parent.type === "TryStatement" && node.parent.finalizer === node;
        ***REMOVED***

        /**
         * Climbs up the tree if the node is not a sentinel node
         * @param {ASTNode***REMOVED*** node node to check.
         * @param {string***REMOVED*** label label of the break or continue statement
         * @returns {boolean***REMOVED*** - return whether the node is a finally block or a sentinel node
         */
        function isInFinallyBlock(node, label) {
            let labelInside = false;
            let sentinelNodeType;

            if (node.type === "BreakStatement" && !node.label) {
                sentinelNodeType = SENTINEL_NODE_TYPE_BREAK;
            ***REMOVED*** else if (node.type === "ContinueStatement") {
                sentinelNodeType = SENTINEL_NODE_TYPE_CONTINUE;
            ***REMOVED*** else {
                sentinelNodeType = SENTINEL_NODE_TYPE_RETURN_THROW;
            ***REMOVED***

            for (
                let currentNode = node;
                currentNode && !sentinelNodeType.test(currentNode.type);
                currentNode = currentNode.parent
            ) {
                if (currentNode.parent.label && label && (currentNode.parent.label.name === label.name)) {
                    labelInside = true;
                ***REMOVED***
                if (isFinallyBlock(currentNode)) {
                    if (label && labelInside) {
                        return false;
                    ***REMOVED***
                    return true;
                ***REMOVED***
            ***REMOVED***
            return false;
        ***REMOVED***

        /**
         * Checks whether the possibly-unsafe statement is inside a finally block.
         * @param {ASTNode***REMOVED*** node node to check.
         * @returns {void***REMOVED***
         */
        function check(node) {
            if (isInFinallyBlock(node, node.label)) {
                context.report({
                    messageId: "unsafeUsage",
                    data: {
                        nodeType: node.type
                    ***REMOVED***,
                    node,
                    line: node.loc.line,
                    column: node.loc.column
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        return {
            ReturnStatement: check,
            ThrowStatement: check,
            BreakStatement: check,
            ContinueStatement: check
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
