// eslint/lib/rules/no-unsafe-optional-chaining.js
/**
 * @fileoverview Rule to disallow unsafe optional chaining
 * @author Yeon JuAn
 */

"use strict";

const UNSAFE_ARITHMETIC_OPERATORS = new Set(["+", "-", "/", "*", "%", "**"]);
const UNSAFE_ASSIGNMENT_OPERATORS = new Set(["+=", "-=", "/=", "*=", "%=", "**="]);
const UNSAFE_RELATIONAL_OPERATORS = new Set(["in", "instanceof"]);

/**
 * Checks whether a node is a destructuring pattern or not
 * @param {ASTNode***REMOVED*** node node to check
 * @returns {boolean***REMOVED*** `true` if a node is a destructuring pattern, otherwise `false`
 */
function isDestructuringPattern(node) {
    return node.type === "ObjectPattern" || node.type === "ArrayPattern";
***REMOVED***

module.exports = {
    meta: {
        type: "problem",

        docs: {
            description: "disallow use of optional chaining in contexts where the `undefined` value is not allowed",
            category: "Possible Errors",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-unsafe-optional-chaining"
        ***REMOVED***,
        schema: [{
            type: "object",
            properties: {
                disallowArithmeticOperators: {
                    type: "boolean",
                    default: false
                ***REMOVED***
            ***REMOVED***,
            additionalProperties: false
        ***REMOVED***],
        fixable: null,
        messages: {
            unsafeOptionalChain: "Unsafe usage of optional chaining. If it short-circuits with 'undefined' the evaluation will throw TypeError.",
            unsafeArithmetic: "Unsafe arithmetic operation on optional chaining. It can result in NaN."
        ***REMOVED***
    ***REMOVED***,

    create(context) {
        const options = context.options[0] || {***REMOVED***;
        const disallowArithmeticOperators = (options.disallowArithmeticOperators) || false;

        /**
         * Reports unsafe usage of optional chaining
         * @param {ASTNode***REMOVED*** node node to report
         * @returns {void***REMOVED***
         */
        function reportUnsafeUsage(node) {
            context.report({
                messageId: "unsafeOptionalChain",
                node
            ***REMOVED***);
        ***REMOVED***

        /**
         * Reports unsafe arithmetic operation on optional chaining
         * @param {ASTNode***REMOVED*** node node to report
         * @returns {void***REMOVED***
         */
        function reportUnsafeArithmetic(node) {
            context.report({
                messageId: "unsafeArithmetic",
                node
            ***REMOVED***);
        ***REMOVED***

        /**
         * Checks and reports if a node can short-circuit with `undefined` by optional chaining.
         * @param {ASTNode***REMOVED*** [node] node to check
         * @param {Function***REMOVED*** reportFunc report function
         * @returns {void***REMOVED***
         */
        function checkUndefinedShortCircuit(node, reportFunc) {
            if (!node) {
                return;
            ***REMOVED***
            switch (node.type) {
                case "LogicalExpression":
                    if (node.operator === "||" || node.operator === "??") {
                        checkUndefinedShortCircuit(node.right, reportFunc);
                    ***REMOVED*** else if (node.operator === "&&") {
                        checkUndefinedShortCircuit(node.left, reportFunc);
                        checkUndefinedShortCircuit(node.right, reportFunc);
                    ***REMOVED***
                    break;
                case "SequenceExpression":
                    checkUndefinedShortCircuit(
                        node.expressions[node.expressions.length - 1],
                        reportFunc
                    );
                    break;
                case "ConditionalExpression":
                    checkUndefinedShortCircuit(node.consequent, reportFunc);
                    checkUndefinedShortCircuit(node.alternate, reportFunc);
                    break;
                case "AwaitExpression":
                    checkUndefinedShortCircuit(node.argument, reportFunc);
                    break;
                case "ChainExpression":
                    reportFunc(node);
                    break;
                default:
                    break;
            ***REMOVED***
        ***REMOVED***

        /**
         * Checks unsafe usage of optional chaining
         * @param {ASTNode***REMOVED*** node node to check
         * @returns {void***REMOVED***
         */
        function checkUnsafeUsage(node) {
            checkUndefinedShortCircuit(node, reportUnsafeUsage);
        ***REMOVED***

        /**
         * Checks unsafe arithmetic operations on optional chaining
         * @param {ASTNode***REMOVED*** node node to check
         * @returns {void***REMOVED***
         */
        function checkUnsafeArithmetic(node) {
            checkUndefinedShortCircuit(node, reportUnsafeArithmetic);
        ***REMOVED***

        return {
            "AssignmentExpression, AssignmentPattern"(node) {
                if (isDestructuringPattern(node.left)) {
                    checkUnsafeUsage(node.right);
                ***REMOVED***
            ***REMOVED***,
            "ClassDeclaration, ClassExpression"(node) {
                checkUnsafeUsage(node.superClass);
            ***REMOVED***,
            CallExpression(node) {
                if (!node.optional) {
                    checkUnsafeUsage(node.callee);
                ***REMOVED***
            ***REMOVED***,
            NewExpression(node) {
                checkUnsafeUsage(node.callee);
            ***REMOVED***,
            VariableDeclarator(node) {
                if (isDestructuringPattern(node.id)) {
                    checkUnsafeUsage(node.init);
                ***REMOVED***
            ***REMOVED***,
            MemberExpression(node) {
                if (!node.optional) {
                    checkUnsafeUsage(node.object);
                ***REMOVED***
            ***REMOVED***,
            TaggedTemplateExpression(node) {
                checkUnsafeUsage(node.tag);
            ***REMOVED***,
            ForOfStatement(node) {
                checkUnsafeUsage(node.right);
            ***REMOVED***,
            SpreadElement(node) {
                if (node.parent && node.parent.type !== "ObjectExpression") {
                    checkUnsafeUsage(node.argument);
                ***REMOVED***
            ***REMOVED***,
            BinaryExpression(node) {
                if (UNSAFE_RELATIONAL_OPERATORS.has(node.operator)) {
                    checkUnsafeUsage(node.right);
                ***REMOVED***
                if (
                    disallowArithmeticOperators &&
                    UNSAFE_ARITHMETIC_OPERATORS.has(node.operator)
                ) {
                    checkUnsafeArithmetic(node.right);
                    checkUnsafeArithmetic(node.left);
                ***REMOVED***
            ***REMOVED***,
            WithStatement(node) {
                checkUnsafeUsage(node.object);
            ***REMOVED***,
            UnaryExpression(node) {
                if (
                    disallowArithmeticOperators &&
                    UNSAFE_ARITHMETIC_OPERATORS.has(node.operator)
                ) {
                    checkUnsafeArithmetic(node.argument);
                ***REMOVED***
            ***REMOVED***,
            AssignmentExpression(node) {
                if (
                    disallowArithmeticOperators &&
                    UNSAFE_ASSIGNMENT_OPERATORS.has(node.operator)
                ) {
                    checkUnsafeArithmetic(node.right);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
