// eslint/lib/rules/no-unused-labels.js
/**
 * @fileoverview Rule to disallow unused labels.
 * @author Toru Nagashima
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "disallow unused labels",
            category: "Best Practices",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-unused-labels"
        ***REMOVED***,

        schema: [],

        fixable: "code",

        messages: {
            unused: "'{{name***REMOVED******REMOVED***:' is defined but never used."
        ***REMOVED***
    ***REMOVED***,

    create(context) {
        const sourceCode = context.getSourceCode();
        let scopeInfo = null;

        /**
         * Adds a scope info to the stack.
         * @param {ASTNode***REMOVED*** node A node to add. This is a LabeledStatement.
         * @returns {void***REMOVED***
         */
        function enterLabeledScope(node) {
            scopeInfo = {
                label: node.label.name,
                used: false,
                upper: scopeInfo
            ***REMOVED***;
        ***REMOVED***

        /**
         * Removes the top of the stack.
         * At the same time, this reports the label if it's never used.
         * @param {ASTNode***REMOVED*** node A node to report. This is a LabeledStatement.
         * @returns {void***REMOVED***
         */
        function exitLabeledScope(node) {
            if (!scopeInfo.used) {
                context.report({
                    node: node.label,
                    messageId: "unused",
                    data: node.label,
                    fix(fixer) {

                        /*
                         * Only perform a fix if there are no comments between the label and the body. This will be the case
                         * when there is exactly one token/comment (the ":") between the label and the body.
                         */
                        if (sourceCode.getTokenAfter(node.label, { includeComments: true ***REMOVED***) ===
                                sourceCode.getTokenBefore(node.body, { includeComments: true ***REMOVED***)) {
                            return fixer.removeRange([node.range[0], node.body.range[0]]);
                        ***REMOVED***

                        return null;
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***

            scopeInfo = scopeInfo.upper;
        ***REMOVED***

        /**
         * Marks the label of a given node as used.
         * @param {ASTNode***REMOVED*** node A node to mark. This is a BreakStatement or
         *      ContinueStatement.
         * @returns {void***REMOVED***
         */
        function markAsUsed(node) {
            if (!node.label) {
                return;
            ***REMOVED***

            const label = node.label.name;
            let info = scopeInfo;

            while (info) {
                if (info.label === label) {
                    info.used = true;
                    break;
                ***REMOVED***
                info = info.upper;
            ***REMOVED***
        ***REMOVED***

        return {
            LabeledStatement: enterLabeledScope,
            "LabeledStatement:exit": exitLabeledScope,
            BreakStatement: markAsUsed,
            ContinueStatement: markAsUsed
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
