// eslint/lib/rules/padded-blocks.js
/**
 * @fileoverview A rule to ensure blank lines within blocks.
 * @author Mathias Schreck <https://github.com/lo1tuma>
 * @deprecated in ESLint v8.53.0
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

/** @type {import('../shared/types').Rule***REMOVED*** */
module.exports = {
    meta: {
        deprecated: true,
        replacedBy: [],
        type: "layout",

        docs: {
            description: "Require or disallow padding within blocks",
            recommended: false,
            url: "https://eslint.org/docs/latest/rules/padded-blocks"
        ***REMOVED***,

        fixable: "whitespace",

        schema: [
            {
                oneOf: [
                    {
                        enum: ["always", "never"]
                    ***REMOVED***,
                    {
                        type: "object",
                        properties: {
                            blocks: {
                                enum: ["always", "never"]
                            ***REMOVED***,
                            switches: {
                                enum: ["always", "never"]
                            ***REMOVED***,
                            classes: {
                                enum: ["always", "never"]
                            ***REMOVED***
                        ***REMOVED***,
                        additionalProperties: false,
                        minProperties: 1
                    ***REMOVED***
                ]
            ***REMOVED***,
            {
                type: "object",
                properties: {
                    allowSingleLineBlocks: {
                        type: "boolean"
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],

        messages: {
            alwaysPadBlock: "Block must be padded by blank lines.",
            neverPadBlock: "Block must not be padded by blank lines."
        ***REMOVED***
    ***REMOVED***,

    create(context) {
        const options = {***REMOVED***;
        const typeOptions = context.options[0] || "always";
        const exceptOptions = context.options[1] || {***REMOVED***;

        if (typeof typeOptions === "string") {
            const shouldHavePadding = typeOptions === "always";

            options.blocks = shouldHavePadding;
            options.switches = shouldHavePadding;
            options.classes = shouldHavePadding;
        ***REMOVED*** else {
            if (Object.prototype.hasOwnProperty.call(typeOptions, "blocks")) {
                options.blocks = typeOptions.blocks === "always";
            ***REMOVED***
            if (Object.prototype.hasOwnProperty.call(typeOptions, "switches")) {
                options.switches = typeOptions.switches === "always";
            ***REMOVED***
            if (Object.prototype.hasOwnProperty.call(typeOptions, "classes")) {
                options.classes = typeOptions.classes === "always";
            ***REMOVED***
        ***REMOVED***

        if (Object.prototype.hasOwnProperty.call(exceptOptions, "allowSingleLineBlocks")) {
            options.allowSingleLineBlocks = exceptOptions.allowSingleLineBlocks === true;
        ***REMOVED***

        const sourceCode = context.sourceCode;

        /**
         * Gets the open brace token from a given node.
         * @param {ASTNode***REMOVED*** node A BlockStatement or SwitchStatement node from which to get the open brace.
         * @returns {Token***REMOVED*** The token of the open brace.
         */
        function getOpenBrace(node) {
            if (node.type === "SwitchStatement") {
                return sourceCode.getTokenBefore(node.cases[0]);
            ***REMOVED***

            if (node.type === "StaticBlock") {
                return sourceCode.getFirstToken(node, { skip: 1 ***REMOVED***); // skip the `static` token
            ***REMOVED***

            // `BlockStatement` or `ClassBody`
            return sourceCode.getFirstToken(node);
        ***REMOVED***

        /**
         * Checks if the given parameter is a comment node
         * @param {ASTNode|Token***REMOVED*** node An AST node or token
         * @returns {boolean***REMOVED*** True if node is a comment
         */
        function isComment(node) {
            return node.type === "Line" || node.type === "Block";
        ***REMOVED***

        /**
         * Checks if there is padding between two tokens
         * @param {Token***REMOVED*** first The first token
         * @param {Token***REMOVED*** second The second token
         * @returns {boolean***REMOVED*** True if there is at least a line between the tokens
         */
        function isPaddingBetweenTokens(first, second) {
            return second.loc.start.line - first.loc.end.line >= 2;
        ***REMOVED***


        /**
         * Checks if the given token has a blank line after it.
         * @param {Token***REMOVED*** token The token to check.
         * @returns {boolean***REMOVED*** Whether or not the token is followed by a blank line.
         */
        function getFirstBlockToken(token) {
            let prev,
                first = token;

            do {
                prev = first;
                first = sourceCode.getTokenAfter(first, { includeComments: true ***REMOVED***);
            ***REMOVED*** while (isComment(first) && first.loc.start.line === prev.loc.end.line);

            return first;
        ***REMOVED***

        /**
         * Checks if the given token is preceded by a blank line.
         * @param {Token***REMOVED*** token The token to check
         * @returns {boolean***REMOVED*** Whether or not the token is preceded by a blank line
         */
        function getLastBlockToken(token) {
            let last = token,
                next;

            do {
                next = last;
                last = sourceCode.getTokenBefore(last, { includeComments: true ***REMOVED***);
            ***REMOVED*** while (isComment(last) && last.loc.end.line === next.loc.start.line);

            return last;
        ***REMOVED***

        /**
         * Checks if a node should be padded, according to the rule config.
         * @param {ASTNode***REMOVED*** node The AST node to check.
         * @throws {Error***REMOVED*** (Unreachable)
         * @returns {boolean***REMOVED*** True if the node should be padded, false otherwise.
         */
        function requirePaddingFor(node) {
            switch (node.type) {
                case "BlockStatement":
                case "StaticBlock":
                    return options.blocks;
                case "SwitchStatement":
                    return options.switches;
                case "ClassBody":
                    return options.classes;

                /* c8 ignore next */
                default:
                    throw new Error("unreachable");
            ***REMOVED***
        ***REMOVED***

        /**
         * Checks the given BlockStatement node to be padded if the block is not empty.
         * @param {ASTNode***REMOVED*** node The AST node of a BlockStatement.
         * @returns {void***REMOVED*** undefined.
         */
        function checkPadding(node) {
            const openBrace = getOpenBrace(node),
                firstBlockToken = getFirstBlockToken(openBrace),
                tokenBeforeFirst = sourceCode.getTokenBefore(firstBlockToken, { includeComments: true ***REMOVED***),
                closeBrace = sourceCode.getLastToken(node),
                lastBlockToken = getLastBlockToken(closeBrace),
                tokenAfterLast = sourceCode.getTokenAfter(lastBlockToken, { includeComments: true ***REMOVED***),
                blockHasTopPadding = isPaddingBetweenTokens(tokenBeforeFirst, firstBlockToken),
                blockHasBottomPadding = isPaddingBetweenTokens(lastBlockToken, tokenAfterLast);

            if (options.allowSingleLineBlocks && astUtils.isTokenOnSameLine(tokenBeforeFirst, tokenAfterLast)) {
                return;
            ***REMOVED***

            if (requirePaddingFor(node)) {

                if (!blockHasTopPadding) {
                    context.report({
                        node,
                        loc: {
                            start: tokenBeforeFirst.loc.start,
                            end: firstBlockToken.loc.start
                        ***REMOVED***,
                        fix(fixer) {
                            return fixer.insertTextAfter(tokenBeforeFirst, "\n");
                        ***REMOVED***,
                        messageId: "alwaysPadBlock"
                    ***REMOVED***);
                ***REMOVED***
                if (!blockHasBottomPadding) {
                    context.report({
                        node,
                        loc: {
                            end: tokenAfterLast.loc.start,
                            start: lastBlockToken.loc.end
                        ***REMOVED***,
                        fix(fixer) {
                            return fixer.insertTextBefore(tokenAfterLast, "\n");
                        ***REMOVED***,
                        messageId: "alwaysPadBlock"
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED*** else {
                if (blockHasTopPadding) {

                    context.report({
                        node,
                        loc: {
                            start: tokenBeforeFirst.loc.start,
                            end: firstBlockToken.loc.start
                        ***REMOVED***,
                        fix(fixer) {
                            return fixer.replaceTextRange([tokenBeforeFirst.range[1], firstBlockToken.range[0] - firstBlockToken.loc.start.column], "\n");
                        ***REMOVED***,
                        messageId: "neverPadBlock"
                    ***REMOVED***);
                ***REMOVED***

                if (blockHasBottomPadding) {

                    context.report({
                        node,
                        loc: {
                            end: tokenAfterLast.loc.start,
                            start: lastBlockToken.loc.end
                        ***REMOVED***,
                        messageId: "neverPadBlock",
                        fix(fixer) {
                            return fixer.replaceTextRange([lastBlockToken.range[1], tokenAfterLast.range[0] - tokenAfterLast.loc.start.column], "\n");
                        ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        const rule = {***REMOVED***;

        if (Object.prototype.hasOwnProperty.call(options, "switches")) {
            rule.SwitchStatement = function(node) {
                if (node.cases.length === 0) {
                    return;
                ***REMOVED***
                checkPadding(node);
            ***REMOVED***;
        ***REMOVED***

        if (Object.prototype.hasOwnProperty.call(options, "blocks")) {
            rule.BlockStatement = function(node) {
                if (node.body.length === 0) {
                    return;
                ***REMOVED***
                checkPadding(node);
            ***REMOVED***;
            rule.StaticBlock = rule.BlockStatement;
        ***REMOVED***

        if (Object.prototype.hasOwnProperty.call(options, "classes")) {
            rule.ClassBody = function(node) {
                if (node.body.length === 0) {
                    return;
                ***REMOVED***
                checkPadding(node);
            ***REMOVED***;
        ***REMOVED***

        return rule;
    ***REMOVED***
***REMOVED***;
