// eslint/lib/rules/require-yield.js
/**
 * @fileoverview Rule to flag the generator functions that does not have yield.
 * @author Toru Nagashima
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "require generator functions to contain `yield`",
            category: "ECMAScript 6",
            recommended: true,
            url: "https://eslint.org/docs/rules/require-yield"
        ***REMOVED***,

        schema: [],

        messages: {
            missingYield: "This generator function does not have 'yield'."
        ***REMOVED***
    ***REMOVED***,

    create(context) {
        const stack = [];

        /**
         * If the node is a generator function, start counting `yield` keywords.
         * @param {Node***REMOVED*** node A function node to check.
         * @returns {void***REMOVED***
         */
        function beginChecking(node) {
            if (node.generator) {
                stack.push(0);
            ***REMOVED***
        ***REMOVED***

        /**
         * If the node is a generator function, end counting `yield` keywords, then
         * reports result.
         * @param {Node***REMOVED*** node A function node to check.
         * @returns {void***REMOVED***
         */
        function endChecking(node) {
            if (!node.generator) {
                return;
            ***REMOVED***

            const countYield = stack.pop();

            if (countYield === 0 && node.body.body.length > 0) {
                context.report({ node, messageId: "missingYield" ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        return {
            FunctionDeclaration: beginChecking,
            "FunctionDeclaration:exit": endChecking,
            FunctionExpression: beginChecking,
            "FunctionExpression:exit": endChecking,

            // Increases the count of `yield` keyword.
            YieldExpression() {

                /* istanbul ignore else */
                if (stack.length > 0) {
                    stack[stack.length - 1] += 1;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
