// eslint/lib/rules/sort-keys.js
/**
 * @fileoverview Rule to require object keys to be sorted
 * @author Toru Nagashima
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils"),
    naturalCompare = require("natural-compare");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Gets the property name of the given `Property` node.
 *
 * - If the property's key is an `Identifier` node, this returns the key's name
 *   whether it's a computed property or not.
 * - If the property has a static name, this returns the static name.
 * - Otherwise, this returns null.
 * @param {ASTNode***REMOVED*** node The `Property` node to get.
 * @returns {string|null***REMOVED*** The property name or null.
 * @private
 */
function getPropertyName(node) {
    const staticName = astUtils.getStaticPropertyName(node);

    if (staticName !== null) {
        return staticName;
    ***REMOVED***

    return node.key.name || null;
***REMOVED***

/**
 * Functions which check that the given 2 names are in specific order.
 *
 * Postfix `I` is meant insensitive.
 * Postfix `N` is meant natural.
 * @private
 */
const isValidOrders = {
    asc(a, b) {
        return a <= b;
    ***REMOVED***,
    ascI(a, b) {
        return a.toLowerCase() <= b.toLowerCase();
    ***REMOVED***,
    ascN(a, b) {
        return naturalCompare(a, b) <= 0;
    ***REMOVED***,
    ascIN(a, b) {
        return naturalCompare(a.toLowerCase(), b.toLowerCase()) <= 0;
    ***REMOVED***,
    desc(a, b) {
        return isValidOrders.asc(b, a);
    ***REMOVED***,
    descI(a, b) {
        return isValidOrders.ascI(b, a);
    ***REMOVED***,
    descN(a, b) {
        return isValidOrders.ascN(b, a);
    ***REMOVED***,
    descIN(a, b) {
        return isValidOrders.ascIN(b, a);
    ***REMOVED***
***REMOVED***;

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "require object keys to be sorted",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/sort-keys"
        ***REMOVED***,

        schema: [
            {
                enum: ["asc", "desc"]
            ***REMOVED***,
            {
                type: "object",
                properties: {
                    caseSensitive: {
                        type: "boolean",
                        default: true
                    ***REMOVED***,
                    natural: {
                        type: "boolean",
                        default: false
                    ***REMOVED***,
                    minKeys: {
                        type: "integer",
                        minimum: 2,
                        default: 2
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],

        messages: {
            sortKeys: "Expected object keys to be in {{natural***REMOVED******REMOVED***{{insensitive***REMOVED******REMOVED***{{order***REMOVED******REMOVED***ending order. '{{thisName***REMOVED******REMOVED***' should be before '{{prevName***REMOVED******REMOVED***'."
        ***REMOVED***
    ***REMOVED***,

    create(context) {

        // Parse options.
        const order = context.options[0] || "asc";
        const options = context.options[1];
        const insensitive = options && options.caseSensitive === false;
        const natural = options && options.natural;
        const minKeys = options && options.minKeys;
        const isValidOrder = isValidOrders[
            order + (insensitive ? "I" : "") + (natural ? "N" : "")
        ];

        // The stack to save the previous property's name for each object literals.
        let stack = null;

        return {
            ObjectExpression(node) {
                stack = {
                    upper: stack,
                    prevName: null,
                    numKeys: node.properties.length
                ***REMOVED***;
            ***REMOVED***,

            "ObjectExpression:exit"() {
                stack = stack.upper;
            ***REMOVED***,

            SpreadElement(node) {
                if (node.parent.type === "ObjectExpression") {
                    stack.prevName = null;
                ***REMOVED***
            ***REMOVED***,

            Property(node) {
                if (node.parent.type === "ObjectPattern") {
                    return;
                ***REMOVED***

                const prevName = stack.prevName;
                const numKeys = stack.numKeys;
                const thisName = getPropertyName(node);

                if (thisName !== null) {
                    stack.prevName = thisName;
                ***REMOVED***

                if (prevName === null || thisName === null || numKeys < minKeys) {
                    return;
                ***REMOVED***

                if (!isValidOrder(prevName, thisName)) {
                    context.report({
                        node,
                        loc: node.key.loc,
                        messageId: "sortKeys",
                        data: {
                            thisName,
                            prevName,
                            order,
                            insensitive: insensitive ? "insensitive " : "",
                            natural: natural ? "natural " : ""
                        ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
