// eslint/lib/rules/space-in-parens.js
/**
 * @fileoverview Disallows or enforces spaces inside of parentheses.
 * @author Jonathan Rajavuori
 */
"use strict";

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "layout",

        docs: {
            description: "enforce consistent spacing inside parentheses",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/space-in-parens"
        ***REMOVED***,

        fixable: "whitespace",

        schema: [
            {
                enum: ["always", "never"]
            ***REMOVED***,
            {
                type: "object",
                properties: {
                    exceptions: {
                        type: "array",
                        items: {
                            enum: ["{***REMOVED***", "[]", "()", "empty"]
                        ***REMOVED***,
                        uniqueItems: true
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],

        messages: {
            missingOpeningSpace: "There must be a space after this paren.",
            missingClosingSpace: "There must be a space before this paren.",
            rejectedOpeningSpace: "There should be no space after this paren.",
            rejectedClosingSpace: "There should be no space before this paren."
        ***REMOVED***
    ***REMOVED***,

    create(context) {
        const ALWAYS = context.options[0] === "always",
            exceptionsArrayOptions = (context.options[1] && context.options[1].exceptions) || [],
            options = {***REMOVED***;

        let exceptions;

        if (exceptionsArrayOptions.length) {
            options.braceException = exceptionsArrayOptions.includes("{***REMOVED***");
            options.bracketException = exceptionsArrayOptions.includes("[]");
            options.parenException = exceptionsArrayOptions.includes("()");
            options.empty = exceptionsArrayOptions.includes("empty");
        ***REMOVED***

        /**
         * Produces an object with the opener and closer exception values
         * @returns {Object***REMOVED*** `openers` and `closers` exception values
         * @private
         */
        function getExceptions() {
            const openers = [],
                closers = [];

            if (options.braceException) {
                openers.push("{");
                closers.push("***REMOVED***");
            ***REMOVED***

            if (options.bracketException) {
                openers.push("[");
                closers.push("]");
            ***REMOVED***

            if (options.parenException) {
                openers.push("(");
                closers.push(")");
            ***REMOVED***

            if (options.empty) {
                openers.push(")");
                closers.push("(");
            ***REMOVED***

            return {
                openers,
                closers
            ***REMOVED***;
        ***REMOVED***

        //--------------------------------------------------------------------------
        // Helpers
        //--------------------------------------------------------------------------
        const sourceCode = context.getSourceCode();

        /**
         * Determines if a token is one of the exceptions for the opener paren
         * @param {Object***REMOVED*** token The token to check
         * @returns {boolean***REMOVED*** True if the token is one of the exceptions for the opener paren
         */
        function isOpenerException(token) {
            return exceptions.openers.includes(token.value);
        ***REMOVED***

        /**
         * Determines if a token is one of the exceptions for the closer paren
         * @param {Object***REMOVED*** token The token to check
         * @returns {boolean***REMOVED*** True if the token is one of the exceptions for the closer paren
         */
        function isCloserException(token) {
            return exceptions.closers.includes(token.value);
        ***REMOVED***

        /**
         * Determines if an opening paren is immediately followed by a required space
         * @param {Object***REMOVED*** openingParenToken The paren token
         * @param {Object***REMOVED*** tokenAfterOpeningParen The token after it
         * @returns {boolean***REMOVED*** True if the opening paren is missing a required space
         */
        function openerMissingSpace(openingParenToken, tokenAfterOpeningParen) {
            if (sourceCode.isSpaceBetweenTokens(openingParenToken, tokenAfterOpeningParen)) {
                return false;
            ***REMOVED***

            if (!options.empty && astUtils.isClosingParenToken(tokenAfterOpeningParen)) {
                return false;
            ***REMOVED***

            if (ALWAYS) {
                return !isOpenerException(tokenAfterOpeningParen);
            ***REMOVED***
            return isOpenerException(tokenAfterOpeningParen);
        ***REMOVED***

        /**
         * Determines if an opening paren is immediately followed by a disallowed space
         * @param {Object***REMOVED*** openingParenToken The paren token
         * @param {Object***REMOVED*** tokenAfterOpeningParen The token after it
         * @returns {boolean***REMOVED*** True if the opening paren has a disallowed space
         */
        function openerRejectsSpace(openingParenToken, tokenAfterOpeningParen) {
            if (!astUtils.isTokenOnSameLine(openingParenToken, tokenAfterOpeningParen)) {
                return false;
            ***REMOVED***

            if (tokenAfterOpeningParen.type === "Line") {
                return false;
            ***REMOVED***

            if (!sourceCode.isSpaceBetweenTokens(openingParenToken, tokenAfterOpeningParen)) {
                return false;
            ***REMOVED***

            if (ALWAYS) {
                return isOpenerException(tokenAfterOpeningParen);
            ***REMOVED***
            return !isOpenerException(tokenAfterOpeningParen);
        ***REMOVED***

        /**
         * Determines if a closing paren is immediately preceded by a required space
         * @param {Object***REMOVED*** tokenBeforeClosingParen The token before the paren
         * @param {Object***REMOVED*** closingParenToken The paren token
         * @returns {boolean***REMOVED*** True if the closing paren is missing a required space
         */
        function closerMissingSpace(tokenBeforeClosingParen, closingParenToken) {
            if (sourceCode.isSpaceBetweenTokens(tokenBeforeClosingParen, closingParenToken)) {
                return false;
            ***REMOVED***

            if (!options.empty && astUtils.isOpeningParenToken(tokenBeforeClosingParen)) {
                return false;
            ***REMOVED***

            if (ALWAYS) {
                return !isCloserException(tokenBeforeClosingParen);
            ***REMOVED***
            return isCloserException(tokenBeforeClosingParen);
        ***REMOVED***

        /**
         * Determines if a closer paren is immediately preceded by a disallowed space
         * @param {Object***REMOVED*** tokenBeforeClosingParen The token before the paren
         * @param {Object***REMOVED*** closingParenToken The paren token
         * @returns {boolean***REMOVED*** True if the closing paren has a disallowed space
         */
        function closerRejectsSpace(tokenBeforeClosingParen, closingParenToken) {
            if (!astUtils.isTokenOnSameLine(tokenBeforeClosingParen, closingParenToken)) {
                return false;
            ***REMOVED***

            if (!sourceCode.isSpaceBetweenTokens(tokenBeforeClosingParen, closingParenToken)) {
                return false;
            ***REMOVED***

            if (ALWAYS) {
                return isCloserException(tokenBeforeClosingParen);
            ***REMOVED***
            return !isCloserException(tokenBeforeClosingParen);
        ***REMOVED***

        //--------------------------------------------------------------------------
        // Public
        //--------------------------------------------------------------------------

        return {
            Program: function checkParenSpaces(node) {
                exceptions = getExceptions();
                const tokens = sourceCode.tokensAndComments;

                tokens.forEach((token, i) => {
                    const prevToken = tokens[i - 1];
                    const nextToken = tokens[i + 1];

                    // if token is not an opening or closing paren token, do nothing
                    if (!astUtils.isOpeningParenToken(token) && !astUtils.isClosingParenToken(token)) {
                        return;
                    ***REMOVED***

                    // if token is an opening paren and is not followed by a required space
                    if (token.value === "(" && openerMissingSpace(token, nextToken)) {
                        context.report({
                            node,
                            loc: token.loc,
                            messageId: "missingOpeningSpace",
                            fix(fixer) {
                                return fixer.insertTextAfter(token, " ");
                            ***REMOVED***
                        ***REMOVED***);
                    ***REMOVED***

                    // if token is an opening paren and is followed by a disallowed space
                    if (token.value === "(" && openerRejectsSpace(token, nextToken)) {
                        context.report({
                            node,
                            loc: { start: token.loc.end, end: nextToken.loc.start ***REMOVED***,
                            messageId: "rejectedOpeningSpace",
                            fix(fixer) {
                                return fixer.removeRange([token.range[1], nextToken.range[0]]);
                            ***REMOVED***
                        ***REMOVED***);
                    ***REMOVED***

                    // if token is a closing paren and is not preceded by a required space
                    if (token.value === ")" && closerMissingSpace(prevToken, token)) {
                        context.report({
                            node,
                            loc: token.loc,
                            messageId: "missingClosingSpace",
                            fix(fixer) {
                                return fixer.insertTextBefore(token, " ");
                            ***REMOVED***
                        ***REMOVED***);
                    ***REMOVED***

                    // if token is a closing paren and is preceded by a disallowed space
                    if (token.value === ")" && closerRejectsSpace(prevToken, token)) {
                        context.report({
                            node,
                            loc: { start: prevToken.loc.end, end: token.loc.start ***REMOVED***,
                            messageId: "rejectedClosingSpace",
                            fix(fixer) {
                                return fixer.removeRange([prevToken.range[1], token.range[0]]);
                            ***REMOVED***
                        ***REMOVED***);
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
