// eslint/lib/rules/vars-on-top.js
/**
 * @fileoverview Rule to enforce var declarations are only at the top of a function.
 * @author Danny Fritz
 * @author Gyandeep Singh
 */
"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "require `var` declarations be placed at the top of their containing scope",
            category: "Best Practices",
            recommended: false,
            url: "https://eslint.org/docs/rules/vars-on-top"
        ***REMOVED***,

        schema: [],
        messages: {
            top: "All 'var' declarations must be at the top of the function scope."
        ***REMOVED***
    ***REMOVED***,

    create(context) {

        //--------------------------------------------------------------------------
        // Helpers
        //--------------------------------------------------------------------------

        // eslint-disable-next-line jsdoc/require-description
        /**
         * @param {ASTNode***REMOVED*** node any node
         * @returns {boolean***REMOVED*** whether the given node structurally represents a directive
         */
        function looksLikeDirective(node) {
            return node.type === "ExpressionStatement" &&
                node.expression.type === "Literal" && typeof node.expression.value === "string";
        ***REMOVED***

        /**
         * Check to see if its a ES6 import declaration
         * @param {ASTNode***REMOVED*** node any node
         * @returns {boolean***REMOVED*** whether the given node represents a import declaration
         */
        function looksLikeImport(node) {
            return node.type === "ImportDeclaration" || node.type === "ImportSpecifier" ||
                node.type === "ImportDefaultSpecifier" || node.type === "ImportNamespaceSpecifier";
        ***REMOVED***

        /**
         * Checks whether a given node is a variable declaration or not.
         * @param {ASTNode***REMOVED*** node any node
         * @returns {boolean***REMOVED*** `true` if the node is a variable declaration.
         */
        function isVariableDeclaration(node) {
            return (
                node.type === "VariableDeclaration" ||
                (
                    node.type === "ExportNamedDeclaration" &&
                    node.declaration &&
                    node.declaration.type === "VariableDeclaration"
                )
            );
        ***REMOVED***

        /**
         * Checks whether this variable is on top of the block body
         * @param {ASTNode***REMOVED*** node The node to check
         * @param {ASTNode[]***REMOVED*** statements collection of ASTNodes for the parent node block
         * @returns {boolean***REMOVED*** True if var is on top otherwise false
         */
        function isVarOnTop(node, statements) {
            const l = statements.length;
            let i = 0;

            // skip over directives
            for (; i < l; ++i) {
                if (!looksLikeDirective(statements[i]) && !looksLikeImport(statements[i])) {
                    break;
                ***REMOVED***
            ***REMOVED***

            for (; i < l; ++i) {
                if (!isVariableDeclaration(statements[i])) {
                    return false;
                ***REMOVED***
                if (statements[i] === node) {
                    return true;
                ***REMOVED***
            ***REMOVED***

            return false;
        ***REMOVED***

        /**
         * Checks whether variable is on top at the global level
         * @param {ASTNode***REMOVED*** node The node to check
         * @param {ASTNode***REMOVED*** parent Parent of the node
         * @returns {void***REMOVED***
         */
        function globalVarCheck(node, parent) {
            if (!isVarOnTop(node, parent.body)) {
                context.report({ node, messageId: "top" ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        /**
         * Checks whether variable is on top at functional block scope level
         * @param {ASTNode***REMOVED*** node The node to check
         * @param {ASTNode***REMOVED*** parent Parent of the node
         * @param {ASTNode***REMOVED*** grandParent Parent of the node's parent
         * @returns {void***REMOVED***
         */
        function blockScopeVarCheck(node, parent, grandParent) {
            if (!(/Function/u.test(grandParent.type) &&
                    parent.type === "BlockStatement" &&
                    isVarOnTop(node, parent.body))) {
                context.report({ node, messageId: "top" ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        //--------------------------------------------------------------------------
        // Public API
        //--------------------------------------------------------------------------

        return {
            "VariableDeclaration[kind='var']"(node) {
                if (node.parent.type === "ExportNamedDeclaration") {
                    globalVarCheck(node.parent, node.parent.parent);
                ***REMOVED*** else if (node.parent.type === "Program") {
                    globalVarCheck(node, node.parent);
                ***REMOVED*** else {
                    blockScopeVarCheck(node, node.parent, node.parent.parent);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
