// axios/lib/helpers/validator.js
'use strict';

var pkg = require('./../../package.json');

var validators = {***REMOVED***;

// eslint-disable-next-line func-names
['object', 'boolean', 'number', 'function', 'string', 'symbol'].forEach(function(type, i) {
  validators[type] = function validator(thing) {
    return typeof thing === type || 'a' + (i < 1 ? 'n ' : ' ') + type;
  ***REMOVED***;
***REMOVED***);

var deprecatedWarnings = {***REMOVED***;
var currentVerArr = pkg.version.split('.');

/**
 * Compare package versions
 * @param {string***REMOVED*** version
 * @param {string?***REMOVED*** thanVersion
 * @returns {boolean***REMOVED***
 */
function isOlderVersion(version, thanVersion) {
  var pkgVersionArr = thanVersion ? thanVersion.split('.') : currentVerArr;
  var destVer = version.split('.');
  for (var i = 0; i < 3; i++) {
    if (pkgVersionArr[i] > destVer[i]) {
      return true;
    ***REMOVED*** else if (pkgVersionArr[i] < destVer[i]) {
      return false;
    ***REMOVED***
  ***REMOVED***
  return false;
***REMOVED***

/**
 * Transitional option validator
 * @param {function|boolean?***REMOVED*** validator
 * @param {string?***REMOVED*** version
 * @param {string***REMOVED*** message
 * @returns {function***REMOVED***
 */
validators.transitional = function transitional(validator, version, message) {
  var isDeprecated = version && isOlderVersion(version);

  function formatMessage(opt, desc) {
    return '[Axios v' + pkg.version + '] Transitional option \'' + opt + '\'' + desc + (message ? '. ' + message : '');
  ***REMOVED***

  // eslint-disable-next-line func-names
  return function(value, opt, opts) {
    if (validator === false) {
      throw new Error(formatMessage(opt, ' has been removed in ' + version));
    ***REMOVED***

    if (isDeprecated && !deprecatedWarnings[opt]) {
      deprecatedWarnings[opt] = true;
      // eslint-disable-next-line no-console
      console.warn(
        formatMessage(
          opt,
          ' has been deprecated since v' + version + ' and will be removed in the near future'
        )
      );
    ***REMOVED***

    return validator ? validator(value, opt, opts) : true;
  ***REMOVED***;
***REMOVED***;

/**
 * Assert object's properties type
 * @param {object***REMOVED*** options
 * @param {object***REMOVED*** schema
 * @param {boolean?***REMOVED*** allowUnknown
 */

function assertOptions(options, schema, allowUnknown) {
  if (typeof options !== 'object') {
    throw new TypeError('options must be an object');
  ***REMOVED***
  var keys = Object.keys(options);
  var i = keys.length;
  while (i-- > 0) {
    var opt = keys[i];
    var validator = schema[opt];
    if (validator) {
      var value = options[opt];
      var result = value === undefined || validator(value, opt, options);
      if (result !== true) {
        throw new TypeError('option ' + opt + ' must be ' + result);
      ***REMOVED***
      continue;
    ***REMOVED***
    if (allowUnknown !== true) {
      throw Error('Unknown option ' + opt);
    ***REMOVED***
  ***REMOVED***
***REMOVED***

module.exports = {
  isOlderVersion: isOlderVersion,
  assertOptions: assertOptions,
  validators: validators
***REMOVED***;
