// jsdom/lib/jsdom/living/traversal/TreeWalker-impl.js
"use strict";

const DOMException = require("domexception/webidl2js-wrapper");
const { filter, FILTER_ACCEPT, FILTER_REJECT, FILTER_SKIP ***REMOVED*** = require("./helpers");

const FIRST = false;
const LAST = true;
const NEXT = false;
const PREVIOUS = true;

exports.implementation = class TreeWalkerImpl {
  constructor(globalObject, args, privateData) {
    this._active = false;
    this.root = privateData.root;
    this.currentNode = this.root;
    this.whatToShow = privateData.whatToShow;
    this.filter = privateData.filter;

    this._globalObject = globalObject;
  ***REMOVED***

  get currentNode() {
    return this._currentNode;
  ***REMOVED***

  set currentNode(node) {
    if (node === null) {
      throw DOMException.create(this._globalObject, ["Cannot set currentNode to null", "NotSupportedError"]);
    ***REMOVED***

    this._currentNode = node;
  ***REMOVED***

  parentNode() {
    let node = this._currentNode;
    while (node !== null && node !== this.root) {
      node = node.parentNode;

      if (node !== null && filter(this, node) === FILTER_ACCEPT) {
        return (this._currentNode = node);
      ***REMOVED***
    ***REMOVED***
    return null;
  ***REMOVED***

  firstChild() {
    return this._traverseChildren(FIRST);
  ***REMOVED***

  lastChild() {
    return this._traverseChildren(LAST);
  ***REMOVED***

  previousSibling() {
    return this._traverseSiblings(PREVIOUS);
  ***REMOVED***

  nextSibling() {
    return this._traverseSiblings(NEXT);
  ***REMOVED***

  previousNode() {
    let node = this._currentNode;

    while (node !== this.root) {
      let sibling = node.previousSibling;

      while (sibling !== null) {
        node = sibling;
        let result = filter(this, node);

        while (result !== FILTER_REJECT && node.hasChildNodes()) {
          node = node.lastChild;
          result = filter(this, node);
        ***REMOVED***

        if (result === FILTER_ACCEPT) {
          return (this._currentNode = node);
        ***REMOVED***

        sibling = node.previousSibling;
      ***REMOVED***

      if (node === this.root || node.parentNode === null) {
        return null;
      ***REMOVED***

      node = node.parentNode;

      if (filter(this, node) === FILTER_ACCEPT) {
        return (this._currentNode = node);
      ***REMOVED***
    ***REMOVED***

    return null;
  ***REMOVED***

  nextNode() {
    let node = this._currentNode;
    let result = FILTER_ACCEPT;

    for (;;) {
      while (result !== FILTER_REJECT && node.hasChildNodes()) {
        node = node.firstChild;
        result = filter(this, node);
        if (result === FILTER_ACCEPT) {
          return (this._currentNode = node);
        ***REMOVED***
      ***REMOVED***

      do {
        if (node === this.root) {
          return null;
        ***REMOVED***

        const sibling = node.nextSibling;

        if (sibling !== null) {
          node = sibling;
          break;
        ***REMOVED***

        node = node.parentNode;
      ***REMOVED*** while (node !== null);

      if (node === null) {
        return null;
      ***REMOVED***

      result = filter(this, node);

      if (result === FILTER_ACCEPT) {
        return (this._currentNode = node);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  _traverseChildren(type) {
    let node = this._currentNode;
    node = type === FIRST ? node.firstChild : node.lastChild;

    if (node === null) {
      return null;
    ***REMOVED***

    main: for (;;) {
      const result = filter(this, node);

      if (result === FILTER_ACCEPT) {
        return (this._currentNode = node);
      ***REMOVED***

      if (result === FILTER_SKIP) {
        const child = type === FIRST ? node.firstChild : node.lastChild;

        if (child !== null) {
          node = child;
          continue;
        ***REMOVED***
      ***REMOVED***

      for (;;) {
        const sibling = type === FIRST ? node.nextSibling : node.previousSibling;

        if (sibling !== null) {
          node = sibling;
          continue main;
        ***REMOVED***

        const parent = node.parentNode;

        if (parent === null || parent === this.root || parent === this._currentNode) {
          return null;
        ***REMOVED***

        node = parent;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  _traverseSiblings(type) {
    let node = this._currentNode;

    if (node === this.root) {
      return null;
    ***REMOVED***

    for (;;) {
      let sibling = type === NEXT ? node.nextSibling : node.previousSibling;

      while (sibling !== null) {
        node = sibling;
        const result = filter(this, node);

        if (result === FILTER_ACCEPT) {
          return (this._currentNode = node);
        ***REMOVED***

        sibling = type === NEXT ? node.firstChild : node.lastChild;

        if (result === FILTER_REJECT || sibling === null) {
          sibling = type === NEXT ? node.nextSibling : node.previousSibling;
        ***REMOVED***
      ***REMOVED***

      node = node.parentNode;

      if (node === null || node === this.root) {
        return null;
      ***REMOVED***

      if (filter(this, node) === FILTER_ACCEPT) {
        return null;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***;
