// jackspeak/dist/commonjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod ***REMOVED***;
***REMOVED***;
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.jack = exports.Jack = exports.isConfigOption = exports.isConfigType = void 0;
const node_util_1 = require("node:util");
const parse_args_js_1 = require("./parse-args.js");
// it's a tiny API, just cast it inline, it's fine
//@ts-ignore
const cliui_1 = __importDefault(require("@isaacs/cliui"));
const node_path_1 = require("node:path");
const width = Math.min((process && process.stdout && process.stdout.columns) || 80, 80);
// indentation spaces from heading level
const indent = (n) => (n - 1) * 2;
const toEnvKey = (pref, key) => {
    return [pref, key.replace(/[^a-zA-Z0-9]+/g, ' ')]
        .join(' ')
        .trim()
        .toUpperCase()
        .replace(/ /g, '_');
***REMOVED***;
const toEnvVal = (value, delim = '\n') => {
    const str = typeof value === 'string' ? value
        : typeof value === 'boolean' ?
            value ? '1'
                : '0'
            : typeof value === 'number' ? String(value)
                : Array.isArray(value) ?
                    value.map((v) => toEnvVal(v)).join(delim)
                    : /* c8 ignore start */ undefined;
    if (typeof str !== 'string') {
        throw new Error(`could not serialize value to environment: ${JSON.stringify(value)***REMOVED***`);
    ***REMOVED***
    /* c8 ignore stop */
    return str;
***REMOVED***;
const fromEnvVal = (env, type, multiple, delim = '\n') => (multiple ?
    env ? env.split(delim).map(v => fromEnvVal(v, type, false))
        : []
    : type === 'string' ? env
        : type === 'boolean' ? env === '1'
            : +env.trim());
const isConfigType = (t) => typeof t === 'string' &&
    (t === 'string' || t === 'number' || t === 'boolean');
exports.isConfigType = isConfigType;
const undefOrType = (v, t) => v === undefined || typeof v === t;
const undefOrTypeArray = (v, t) => v === undefined || (Array.isArray(v) && v.every(x => typeof x === t));
const isValidOption = (v, vo) => Array.isArray(v) ? v.every(x => isValidOption(x, vo)) : vo.includes(v);
// print the value type, for error message reporting
const valueType = (v) => typeof v === 'string' ? 'string'
    : typeof v === 'boolean' ? 'boolean'
        : typeof v === 'number' ? 'number'
            : Array.isArray(v) ?
                joinTypes([...new Set(v.map(v => valueType(v)))]) + '[]'
                : `${v.type***REMOVED***${v.multiple ? '[]' : ''***REMOVED***`;
const joinTypes = (types) => types.length === 1 && typeof types[0] === 'string' ?
    types[0]
    : `(${types.join('|')***REMOVED***)`;
const isValidValue = (v, type, multi) => {
    if (multi) {
        if (!Array.isArray(v))
            return false;
        return !v.some((v) => !isValidValue(v, type, false));
    ***REMOVED***
    if (Array.isArray(v))
        return false;
    return typeof v === type;
***REMOVED***;
const isConfigOption = (o, type, multi) => !!o &&
    typeof o === 'object' &&
    (0, exports.isConfigType)(o.type) &&
    o.type === type &&
    undefOrType(o.short, 'string') &&
    undefOrType(o.description, 'string') &&
    undefOrType(o.hint, 'string') &&
    undefOrType(o.validate, 'function') &&
    (o.type === 'boolean' ?
        o.validOptions === undefined
        : undefOrTypeArray(o.validOptions, o.type)) &&
    (o.default === undefined || isValidValue(o.default, type, multi)) &&
    !!o.multiple === multi;
exports.isConfigOption = isConfigOption;
function num(o = {***REMOVED***) {
    const { default: def, validate: val, validOptions, ...rest ***REMOVED*** = o;
    if (def !== undefined && !isValidValue(def, 'number', false)) {
        throw new TypeError('invalid default value', {
            cause: {
                found: def,
                wanted: 'number',
            ***REMOVED***,
        ***REMOVED***);
    ***REMOVED***
    if (!undefOrTypeArray(validOptions, 'number')) {
        throw new TypeError('invalid validOptions', {
            cause: {
                found: validOptions,
                wanted: 'number[]',
            ***REMOVED***,
        ***REMOVED***);
    ***REMOVED***
    const validate = val ?
        val
        : undefined;
    return {
        ...rest,
        default: def,
        validate,
        validOptions,
        type: 'number',
        multiple: false,
    ***REMOVED***;
***REMOVED***
function numList(o = {***REMOVED***) {
    const { default: def, validate: val, validOptions, ...rest ***REMOVED*** = o;
    if (def !== undefined && !isValidValue(def, 'number', true)) {
        throw new TypeError('invalid default value', {
            cause: {
                found: def,
                wanted: 'number[]',
            ***REMOVED***,
        ***REMOVED***);
    ***REMOVED***
    if (!undefOrTypeArray(validOptions, 'number')) {
        throw new TypeError('invalid validOptions', {
            cause: {
                found: validOptions,
                wanted: 'number[]',
            ***REMOVED***,
        ***REMOVED***);
    ***REMOVED***
    const validate = val ?
        val
        : undefined;
    return {
        ...rest,
        default: def,
        validate,
        validOptions,
        type: 'number',
        multiple: true,
    ***REMOVED***;
***REMOVED***
function opt(o = {***REMOVED***) {
    const { default: def, validate: val, validOptions, ...rest ***REMOVED*** = o;
    if (def !== undefined && !isValidValue(def, 'string', false)) {
        throw new TypeError('invalid default value', {
            cause: {
                found: def,
                wanted: 'string',
            ***REMOVED***,
        ***REMOVED***);
    ***REMOVED***
    if (!undefOrTypeArray(validOptions, 'string')) {
        throw new TypeError('invalid validOptions', {
            cause: {
                found: validOptions,
                wanted: 'string[]',
            ***REMOVED***,
        ***REMOVED***);
    ***REMOVED***
    const validate = val ?
        val
        : undefined;
    return {
        ...rest,
        default: def,
        validate,
        validOptions,
        type: 'string',
        multiple: false,
    ***REMOVED***;
***REMOVED***
function optList(o = {***REMOVED***) {
    const { default: def, validate: val, validOptions, ...rest ***REMOVED*** = o;
    if (def !== undefined && !isValidValue(def, 'string', true)) {
        throw new TypeError('invalid default value', {
            cause: {
                found: def,
                wanted: 'string[]',
            ***REMOVED***,
        ***REMOVED***);
    ***REMOVED***
    if (!undefOrTypeArray(validOptions, 'string')) {
        throw new TypeError('invalid validOptions', {
            cause: {
                found: validOptions,
                wanted: 'string[]',
            ***REMOVED***,
        ***REMOVED***);
    ***REMOVED***
    const validate = val ?
        val
        : undefined;
    return {
        ...rest,
        default: def,
        validate,
        validOptions,
        type: 'string',
        multiple: true,
    ***REMOVED***;
***REMOVED***
function flag(o = {***REMOVED***) {
    const { hint, default: def, validate: val, ...rest ***REMOVED*** = o;
    delete rest.validOptions;
    if (def !== undefined && !isValidValue(def, 'boolean', false)) {
        throw new TypeError('invalid default value');
    ***REMOVED***
    const validate = val ?
        val
        : undefined;
    if (hint !== undefined) {
        throw new TypeError('cannot provide hint for flag');
    ***REMOVED***
    return {
        ...rest,
        default: def,
        validate,
        type: 'boolean',
        multiple: false,
    ***REMOVED***;
***REMOVED***
function flagList(o = {***REMOVED***) {
    const { hint, default: def, validate: val, ...rest ***REMOVED*** = o;
    delete rest.validOptions;
    if (def !== undefined && !isValidValue(def, 'boolean', true)) {
        throw new TypeError('invalid default value');
    ***REMOVED***
    const validate = val ?
        val
        : undefined;
    if (hint !== undefined) {
        throw new TypeError('cannot provide hint for flag list');
    ***REMOVED***
    return {
        ...rest,
        default: def,
        validate,
        type: 'boolean',
        multiple: true,
    ***REMOVED***;
***REMOVED***
const toParseArgsOptionsConfig = (options) => {
    const c = {***REMOVED***;
    for (const longOption in options) {
        const config = options[longOption];
        /* c8 ignore start */
        if (!config) {
            throw new Error('config must be an object: ' + longOption);
        ***REMOVED***
        /* c8 ignore start */
        if ((0, exports.isConfigOption)(config, 'number', true)) {
            c[longOption] = {
                type: 'string',
                multiple: true,
                default: config.default?.map(c => String(c)),
            ***REMOVED***;
        ***REMOVED***
        else if ((0, exports.isConfigOption)(config, 'number', false)) {
            c[longOption] = {
                type: 'string',
                multiple: false,
                default: config.default === undefined ?
                    undefined
                    : String(config.default),
            ***REMOVED***;
        ***REMOVED***
        else {
            const conf = config;
            c[longOption] = {
                type: conf.type,
                multiple: !!conf.multiple,
                default: conf.default,
            ***REMOVED***;
        ***REMOVED***
        const clo = c[longOption];
        if (typeof config.short === 'string') {
            clo.short = config.short;
        ***REMOVED***
        if (config.type === 'boolean' &&
            !longOption.startsWith('no-') &&
            !options[`no-${longOption***REMOVED***`]) {
            c[`no-${longOption***REMOVED***`] = {
                type: 'boolean',
                multiple: config.multiple,
            ***REMOVED***;
        ***REMOVED***
    ***REMOVED***
    return c;
***REMOVED***;
const isHeading = (r) => r.type === 'heading';
const isDescription = (r) => r.type === 'description';
/**
 * Class returned by the {@link jack***REMOVED*** function and all configuration
 * definition methods.  This is what gets chained together.
 */
class Jack {
    #configSet;
    #shorts;
    #options;
    #fields = [];
    #env;
    #envPrefix;
    #allowPositionals;
    #usage;
    #usageMarkdown;
    constructor(options = {***REMOVED***) {
        this.#options = options;
        this.#allowPositionals = options.allowPositionals !== false;
        this.#env =
            this.#options.env === undefined ? process.env : this.#options.env;
        this.#envPrefix = options.envPrefix;
        // We need to fib a little, because it's always the same object, but it
        // starts out as having an empty config set.  Then each method that adds
        // fields returns `this as Jack<C & { ...newConfigs ***REMOVED***>`
        this.#configSet = Object.create(null);
        this.#shorts = Object.create(null);
    ***REMOVED***
    /**
     * Set the default value (which will still be overridden by env or cli)
     * as if from a parsed config file. The optional `source` param, if
     * provided, will be included in error messages if a value is invalid or
     * unknown.
     */
    setConfigValues(values, source = '') {
        try {
            this.validate(values);
        ***REMOVED***
        catch (er) {
            const e = er;
            if (source && e && typeof e === 'object') {
                if (e.cause && typeof e.cause === 'object') {
                    Object.assign(e.cause, { path: source ***REMOVED***);
                ***REMOVED***
                else {
                    e.cause = { path: source ***REMOVED***;
                ***REMOVED***
            ***REMOVED***
            throw e;
        ***REMOVED***
        for (const [field, value] of Object.entries(values)) {
            const my = this.#configSet[field];
            // already validated, just for TS's benefit
            /* c8 ignore start */
            if (!my) {
                throw new Error('unexpected field in config set: ' + field, {
                    cause: { found: field ***REMOVED***,
                ***REMOVED***);
            ***REMOVED***
            /* c8 ignore stop */
            my.default = value;
        ***REMOVED***
        return this;
    ***REMOVED***
    /**
     * Parse a string of arguments, and return the resulting
     * `{ values, positionals ***REMOVED***` object.
     *
     * If an {@link JackOptions#envPrefix***REMOVED*** is set, then it will read default
     * values from the environment, and write the resulting values back
     * to the environment as well.
     *
     * Environment values always take precedence over any other value, except
     * an explicit CLI setting.
     */
    parse(args = process.argv) {
        this.loadEnvDefaults();
        const p = this.parseRaw(args);
        this.applyDefaults(p);
        this.writeEnv(p);
        return p;
    ***REMOVED***
    loadEnvDefaults() {
        if (this.#envPrefix) {
            for (const [field, my] of Object.entries(this.#configSet)) {
                const ek = toEnvKey(this.#envPrefix, field);
                const env = this.#env[ek];
                if (env !== undefined) {
                    my.default = fromEnvVal(env, my.type, !!my.multiple, my.delim);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    applyDefaults(p) {
        for (const [field, c] of Object.entries(this.#configSet)) {
            if (c.default !== undefined && !(field in p.values)) {
                //@ts-ignore
                p.values[field] = c.default;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    /**
     * Only parse the command line arguments passed in.
     * Does not strip off the `node script.js` bits, so it must be just the
     * arguments you wish to have parsed.
     * Does not read from or write to the environment, or set defaults.
     */
    parseRaw(args) {
        if (args === process.argv) {
            args = args.slice(process._eval !== undefined ? 1 : 2);
        ***REMOVED***
        const options = toParseArgsOptionsConfig(this.#configSet);
        const result = (0, parse_args_js_1.parseArgs)({
            args,
            options,
            // always strict, but using our own logic
            strict: false,
            allowPositionals: this.#allowPositionals,
            tokens: true,
        ***REMOVED***);
        const p = {
            values: {***REMOVED***,
            positionals: [],
        ***REMOVED***;
        for (const token of result.tokens) {
            if (token.kind === 'positional') {
                p.positionals.push(token.value);
                if (this.#options.stopAtPositional ||
                    this.#options.stopAtPositionalTest?.(token.value)) {
                    p.positionals.push(...args.slice(token.index + 1));
                    break;
                ***REMOVED***
            ***REMOVED***
            else if (token.kind === 'option') {
                let value = undefined;
                if (token.name.startsWith('no-')) {
                    const my = this.#configSet[token.name];
                    const pname = token.name.substring('no-'.length);
                    const pos = this.#configSet[pname];
                    if (pos &&
                        pos.type === 'boolean' &&
                        (!my ||
                            (my.type === 'boolean' && !!my.multiple === !!pos.multiple))) {
                        value = false;
                        token.name = pname;
                    ***REMOVED***
                ***REMOVED***
                const my = this.#configSet[token.name];
                if (!my) {
                    throw new Error(`Unknown option '${token.rawName***REMOVED***'. ` +
                        `To specify a positional argument starting with a '-', ` +
                        `place it at the end of the command after '--', as in ` +
                        `'-- ${token.rawName***REMOVED***'`, {
                        cause: {
                            found: token.rawName + (token.value ? `=${token.value***REMOVED***` : ''),
                        ***REMOVED***,
                    ***REMOVED***);
                ***REMOVED***
                if (value === undefined) {
                    if (token.value === undefined) {
                        if (my.type !== 'boolean') {
                            throw new Error(`No value provided for ${token.rawName***REMOVED***, expected ${my.type***REMOVED***`, {
                                cause: {
                                    name: token.rawName,
                                    wanted: valueType(my),
                                ***REMOVED***,
                            ***REMOVED***);
                        ***REMOVED***
                        value = true;
                    ***REMOVED***
                    else {
                        if (my.type === 'boolean') {
                            throw new Error(`Flag ${token.rawName***REMOVED*** does not take a value, received '${token.value***REMOVED***'`, { cause: { found: token ***REMOVED*** ***REMOVED***);
                        ***REMOVED***
                        if (my.type === 'string') {
                            value = token.value;
                        ***REMOVED***
                        else {
                            value = +token.value;
                            if (value !== value) {
                                throw new Error(`Invalid value '${token.value***REMOVED***' provided for ` +
                                    `'${token.rawName***REMOVED***' option, expected number`, {
                                    cause: {
                                        name: token.rawName,
                                        found: token.value,
                                        wanted: 'number',
                                    ***REMOVED***,
                                ***REMOVED***);
                            ***REMOVED***
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
                if (my.multiple) {
                    const pv = p.values;
                    const tn = pv[token.name] ?? [];
                    pv[token.name] = tn;
                    tn.push(value);
                ***REMOVED***
                else {
                    const pv = p.values;
                    pv[token.name] = value;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        for (const [field, value] of Object.entries(p.values)) {
            const valid = this.#configSet[field]?.validate;
            const validOptions = this.#configSet[field]?.validOptions;
            let cause;
            if (validOptions && !isValidOption(value, validOptions)) {
                cause = { name: field, found: value, validOptions: validOptions ***REMOVED***;
            ***REMOVED***
            if (valid && !valid(value)) {
                cause = cause || { name: field, found: value ***REMOVED***;
            ***REMOVED***
            if (cause) {
                throw new Error(`Invalid value provided for --${field***REMOVED***: ${JSON.stringify(value)***REMOVED***`, { cause ***REMOVED***);
            ***REMOVED***
        ***REMOVED***
        return p;
    ***REMOVED***
    /**
     * do not set fields as 'no-foo' if 'foo' exists and both are bools
     * just set foo.
     */
    #noNoFields(f, val, s = f) {
        if (!f.startsWith('no-') || typeof val !== 'boolean')
            return;
        const yes = f.substring('no-'.length);
        // recurse so we get the core config key we care about.
        this.#noNoFields(yes, val, s);
        if (this.#configSet[yes]?.type === 'boolean') {
            throw new Error(`do not set '${s***REMOVED***', instead set '${yes***REMOVED***' as desired.`, { cause: { found: s, wanted: yes ***REMOVED*** ***REMOVED***);
        ***REMOVED***
    ***REMOVED***
    /**
     * Validate that any arbitrary object is a valid configuration `values`
     * object.  Useful when loading config files or other sources.
     */
    validate(o) {
        if (!o || typeof o !== 'object') {
            throw new Error('Invalid config: not an object', {
                cause: { found: o ***REMOVED***,
            ***REMOVED***);
        ***REMOVED***
        const opts = o;
        for (const field in o) {
            const value = opts[field];
            /* c8 ignore next - for TS */
            if (value === undefined)
                continue;
            this.#noNoFields(field, value);
            const config = this.#configSet[field];
            if (!config) {
                throw new Error(`Unknown config option: ${field***REMOVED***`, {
                    cause: { found: field ***REMOVED***,
                ***REMOVED***);
            ***REMOVED***
            if (!isValidValue(value, config.type, !!config.multiple)) {
                throw new Error(`Invalid value ${valueType(value)***REMOVED*** for ${field***REMOVED***, expected ${valueType(config)***REMOVED***`, {
                    cause: {
                        name: field,
                        found: value,
                        wanted: valueType(config),
                    ***REMOVED***,
                ***REMOVED***);
            ***REMOVED***
            let cause;
            if (config.validOptions &&
                !isValidOption(value, config.validOptions)) {
                cause = {
                    name: field,
                    found: value,
                    validOptions: config.validOptions,
                ***REMOVED***;
            ***REMOVED***
            if (config.validate && !config.validate(value)) {
                cause = cause || { name: field, found: value ***REMOVED***;
            ***REMOVED***
            if (cause) {
                throw new Error(`Invalid config value for ${field***REMOVED***: ${value***REMOVED***`, {
                    cause,
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    writeEnv(p) {
        if (!this.#env || !this.#envPrefix)
            return;
        for (const [field, value] of Object.entries(p.values)) {
            const my = this.#configSet[field];
            this.#env[toEnvKey(this.#envPrefix, field)] = toEnvVal(value, my?.delim);
        ***REMOVED***
    ***REMOVED***
    /**
     * Add a heading to the usage output banner
     */
    heading(text, level, { pre = false ***REMOVED*** = {***REMOVED***) {
        if (level === undefined) {
            level = this.#fields.some(r => isHeading(r)) ? 2 : 1;
        ***REMOVED***
        this.#fields.push({ type: 'heading', text, level, pre ***REMOVED***);
        return this;
    ***REMOVED***
    /**
     * Add a long-form description to the usage output at this position.
     */
    description(text, { pre ***REMOVED*** = {***REMOVED***) {
        this.#fields.push({ type: 'description', text, pre ***REMOVED***);
        return this;
    ***REMOVED***
    /**
     * Add one or more number fields.
     */
    num(fields) {
        return this.#addFields(fields, num);
    ***REMOVED***
    /**
     * Add one or more multiple number fields.
     */
    numList(fields) {
        return this.#addFields(fields, numList);
    ***REMOVED***
    /**
     * Add one or more string option fields.
     */
    opt(fields) {
        return this.#addFields(fields, opt);
    ***REMOVED***
    /**
     * Add one or more multiple string option fields.
     */
    optList(fields) {
        return this.#addFields(fields, optList);
    ***REMOVED***
    /**
     * Add one or more flag fields.
     */
    flag(fields) {
        return this.#addFields(fields, flag);
    ***REMOVED***
    /**
     * Add one or more multiple flag fields.
     */
    flagList(fields) {
        return this.#addFields(fields, flagList);
    ***REMOVED***
    /**
     * Generic field definition method. Similar to flag/flagList/number/etc,
     * but you must specify the `type` (and optionally `multiple` and `delim`)
     * fields on each one, or Jack won't know how to define them.
     */
    addFields(fields) {
        const next = this;
        for (const [name, field] of Object.entries(fields)) {
            this.#validateName(name, field);
            next.#fields.push({
                type: 'config',
                name,
                value: field,
            ***REMOVED***);
        ***REMOVED***
        Object.assign(next.#configSet, fields);
        return next;
    ***REMOVED***
    #addFields(fields, fn) {
        const next = this;
        Object.assign(next.#configSet, Object.fromEntries(Object.entries(fields).map(([name, field]) => {
            this.#validateName(name, field);
            const option = fn(field);
            next.#fields.push({
                type: 'config',
                name,
                value: option,
            ***REMOVED***);
            return [name, option];
        ***REMOVED***)));
        return next;
    ***REMOVED***
    #validateName(name, field) {
        if (!/^[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?$/.test(name)) {
            throw new TypeError(`Invalid option name: ${name***REMOVED***, ` +
                `must be '-' delimited ASCII alphanumeric`);
        ***REMOVED***
        if (this.#configSet[name]) {
            throw new TypeError(`Cannot redefine option ${field***REMOVED***`);
        ***REMOVED***
        if (this.#shorts[name]) {
            throw new TypeError(`Cannot redefine option ${name***REMOVED***, already ` +
                `in use for ${this.#shorts[name]***REMOVED***`);
        ***REMOVED***
        if (field.short) {
            if (!/^[a-zA-Z0-9]$/.test(field.short)) {
                throw new TypeError(`Invalid ${name***REMOVED*** short option: ${field.short***REMOVED***, ` +
                    'must be 1 ASCII alphanumeric character');
            ***REMOVED***
            if (this.#shorts[field.short]) {
                throw new TypeError(`Invalid ${name***REMOVED*** short option: ${field.short***REMOVED***, ` +
                    `already in use for ${this.#shorts[field.short]***REMOVED***`);
            ***REMOVED***
            this.#shorts[field.short] = name;
            this.#shorts[name] = name;
        ***REMOVED***
    ***REMOVED***
    /**
     * Return the usage banner for the given configuration
     */
    usage() {
        if (this.#usage)
            return this.#usage;
        let headingLevel = 1;
        const ui = (0, cliui_1.default)({ width ***REMOVED***);
        const first = this.#fields[0];
        let start = first?.type === 'heading' ? 1 : 0;
        if (first?.type === 'heading') {
            ui.div({
                padding: [0, 0, 0, 0],
                text: normalize(first.text),
            ***REMOVED***);
        ***REMOVED***
        ui.div({ padding: [0, 0, 0, 0], text: 'Usage:' ***REMOVED***);
        if (this.#options.usage) {
            ui.div({
                text: this.#options.usage,
                padding: [0, 0, 0, 2],
            ***REMOVED***);
        ***REMOVED***
        else {
            const cmd = (0, node_path_1.basename)(String(process.argv[1]));
            const shortFlags = [];
            const shorts = [];
            const flags = [];
            const opts = [];
            for (const [field, config] of Object.entries(this.#configSet)) {
                if (config.short) {
                    if (config.type === 'boolean')
                        shortFlags.push(config.short);
                    else
                        shorts.push([config.short, config.hint || field]);
                ***REMOVED***
                else {
                    if (config.type === 'boolean')
                        flags.push(field);
                    else
                        opts.push([field, config.hint || field]);
                ***REMOVED***
            ***REMOVED***
            const sf = shortFlags.length ? ' -' + shortFlags.join('') : '';
            const so = shorts.map(([k, v]) => ` --${k***REMOVED***=<${v***REMOVED***>`).join('');
            const lf = flags.map(k => ` --${k***REMOVED***`).join('');
            const lo = opts.map(([k, v]) => ` --${k***REMOVED***=<${v***REMOVED***>`).join('');
            const usage = `${cmd***REMOVED***${sf***REMOVED***${so***REMOVED***${lf***REMOVED***${lo***REMOVED***`.trim();
            ui.div({
                text: usage,
                padding: [0, 0, 0, 2],
            ***REMOVED***);
        ***REMOVED***
        ui.div({ padding: [0, 0, 0, 0], text: '' ***REMOVED***);
        const maybeDesc = this.#fields[start];
        if (maybeDesc && isDescription(maybeDesc)) {
            const print = normalize(maybeDesc.text, maybeDesc.pre);
            start++;
            ui.div({ padding: [0, 0, 0, 0], text: print ***REMOVED***);
            ui.div({ padding: [0, 0, 0, 0], text: '' ***REMOVED***);
        ***REMOVED***
        const { rows, maxWidth ***REMOVED*** = this.#usageRows(start);
        // every heading/description after the first gets indented by 2
        // extra spaces.
        for (const row of rows) {
            if (row.left) {
                // If the row is too long, don't wrap it
                // Bump the right-hand side down a line to make room
                const configIndent = indent(Math.max(headingLevel, 2));
                if (row.left.length > maxWidth - 3) {
                    ui.div({ text: row.left, padding: [0, 0, 0, configIndent] ***REMOVED***);
                    ui.div({ text: row.text, padding: [0, 0, 0, maxWidth] ***REMOVED***);
                ***REMOVED***
                else {
                    ui.div({
                        text: row.left,
                        padding: [0, 1, 0, configIndent],
                        width: maxWidth,
                    ***REMOVED***, { padding: [0, 0, 0, 0], text: row.text ***REMOVED***);
                ***REMOVED***
                if (row.skipLine) {
                    ui.div({ padding: [0, 0, 0, 0], text: '' ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
            else {
                if (isHeading(row)) {
                    const { level ***REMOVED*** = row;
                    headingLevel = level;
                    // only h1 and h2 have bottom padding
                    // h3-h6 do not
                    const b = level <= 2 ? 1 : 0;
                    ui.div({ ...row, padding: [0, 0, b, indent(level)] ***REMOVED***);
                ***REMOVED***
                else {
                    ui.div({ ...row, padding: [0, 0, 1, indent(headingLevel + 1)] ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        return (this.#usage = ui.toString());
    ***REMOVED***
    /**
     * Return the usage banner markdown for the given configuration
     */
    usageMarkdown() {
        if (this.#usageMarkdown)
            return this.#usageMarkdown;
        const out = [];
        let headingLevel = 1;
        const first = this.#fields[0];
        let start = first?.type === 'heading' ? 1 : 0;
        if (first?.type === 'heading') {
            out.push(`# ${normalizeOneLine(first.text)***REMOVED***`);
        ***REMOVED***
        out.push('Usage:');
        if (this.#options.usage) {
            out.push(normalizeMarkdown(this.#options.usage, true));
        ***REMOVED***
        else {
            const cmd = (0, node_path_1.basename)(String(process.argv[1]));
            const shortFlags = [];
            const shorts = [];
            const flags = [];
            const opts = [];
            for (const [field, config] of Object.entries(this.#configSet)) {
                if (config.short) {
                    if (config.type === 'boolean')
                        shortFlags.push(config.short);
                    else
                        shorts.push([config.short, config.hint || field]);
                ***REMOVED***
                else {
                    if (config.type === 'boolean')
                        flags.push(field);
                    else
                        opts.push([field, config.hint || field]);
                ***REMOVED***
            ***REMOVED***
            const sf = shortFlags.length ? ' -' + shortFlags.join('') : '';
            const so = shorts.map(([k, v]) => ` --${k***REMOVED***=<${v***REMOVED***>`).join('');
            const lf = flags.map(k => ` --${k***REMOVED***`).join('');
            const lo = opts.map(([k, v]) => ` --${k***REMOVED***=<${v***REMOVED***>`).join('');
            const usage = `${cmd***REMOVED***${sf***REMOVED***${so***REMOVED***${lf***REMOVED***${lo***REMOVED***`.trim();
            out.push(normalizeMarkdown(usage, true));
        ***REMOVED***
        const maybeDesc = this.#fields[start];
        if (maybeDesc && isDescription(maybeDesc)) {
            out.push(normalizeMarkdown(maybeDesc.text, maybeDesc.pre));
            start++;
        ***REMOVED***
        const { rows ***REMOVED*** = this.#usageRows(start);
        // heading level in markdown is number of # ahead of text
        for (const row of rows) {
            if (row.left) {
                out.push('#'.repeat(headingLevel + 1) +
                    ' ' +
                    normalizeOneLine(row.left, true));
                if (row.text)
                    out.push(normalizeMarkdown(row.text));
            ***REMOVED***
            else if (isHeading(row)) {
                const { level ***REMOVED*** = row;
                headingLevel = level;
                out.push(`${'#'.repeat(headingLevel)***REMOVED*** ${normalizeOneLine(row.text, row.pre)***REMOVED***`);
            ***REMOVED***
            else {
                out.push(normalizeMarkdown(row.text, !!row.pre));
            ***REMOVED***
        ***REMOVED***
        return (this.#usageMarkdown = out.join('\n\n') + '\n');
    ***REMOVED***
    #usageRows(start) {
        // turn each config type into a row, and figure out the width of the
        // left hand indentation for the option descriptions.
        let maxMax = Math.max(12, Math.min(26, Math.floor(width / 3)));
        let maxWidth = 8;
        let prev = undefined;
        const rows = [];
        for (const field of this.#fields.slice(start)) {
            if (field.type !== 'config') {
                if (prev?.type === 'config')
                    prev.skipLine = true;
                prev = undefined;
                field.text = normalize(field.text, !!field.pre);
                rows.push(field);
                continue;
            ***REMOVED***
            const { value ***REMOVED*** = field;
            const desc = value.description || '';
            const mult = value.multiple ? 'Can be set multiple times' : '';
            const opts = value.validOptions?.length ?
                `Valid options:${value.validOptions.map(v => ` ${JSON.stringify(v)***REMOVED***`)***REMOVED***`
                : '';
            const dmDelim = desc.includes('\n') ? '\n\n' : '\n';
            const extra = [opts, mult].join(dmDelim).trim();
            const text = (normalize(desc) + dmDelim + extra).trim();
            const hint = value.hint ||
                (value.type === 'number' ? 'n'
                    : value.type === 'string' ? field.name
                        : undefined);
            const short = !value.short ? ''
                : value.type === 'boolean' ? `-${value.short***REMOVED*** `
                    : `-${value.short***REMOVED***<${hint***REMOVED***> `;
            const left = value.type === 'boolean' ?
                `${short***REMOVED***--${field.name***REMOVED***`
                : `${short***REMOVED***--${field.name***REMOVED***=<${hint***REMOVED***>`;
            const row = { text, left, type: 'config' ***REMOVED***;
            if (text.length > width - maxMax) {
                row.skipLine = true;
            ***REMOVED***
            if (prev && left.length > maxMax)
                prev.skipLine = true;
            prev = row;
            const len = left.length + 4;
            if (len > maxWidth && len < maxMax) {
                maxWidth = len;
            ***REMOVED***
            rows.push(row);
        ***REMOVED***
        return { rows, maxWidth ***REMOVED***;
    ***REMOVED***
    /**
     * Return the configuration options as a plain object
     */
    toJSON() {
        return Object.fromEntries(Object.entries(this.#configSet).map(([field, def]) => [
            field,
            {
                type: def.type,
                ...(def.multiple ? { multiple: true ***REMOVED*** : {***REMOVED***),
                ...(def.delim ? { delim: def.delim ***REMOVED*** : {***REMOVED***),
                ...(def.short ? { short: def.short ***REMOVED*** : {***REMOVED***),
                ...(def.description ?
                    { description: normalize(def.description) ***REMOVED***
                    : {***REMOVED***),
                ...(def.validate ? { validate: def.validate ***REMOVED*** : {***REMOVED***),
                ...(def.validOptions ? { validOptions: def.validOptions ***REMOVED*** : {***REMOVED***),
                ...(def.default !== undefined ? { default: def.default ***REMOVED*** : {***REMOVED***),
                ...(def.hint ? { hint: def.hint ***REMOVED*** : {***REMOVED***),
            ***REMOVED***,
        ]));
    ***REMOVED***
    /**
     * Custom printer for `util.inspect`
     */
    [node_util_1.inspect.custom](_, options) {
        return `Jack ${(0, node_util_1.inspect)(this.toJSON(), options)***REMOVED***`;
    ***REMOVED***
***REMOVED***
exports.Jack = Jack;
// Unwrap and un-indent, so we can wrap description
// strings however makes them look nice in the code.
const normalize = (s, pre = false) => {
    if (pre)
        // prepend a ZWSP to each line so cliui doesn't strip it.
        return s
            .split('\n')
            .map(l => `\u200b${l***REMOVED***`)
            .join('\n');
    return s
        .split(/^\s*```\s*$/gm)
        .map((s, i) => {
        if (i % 2 === 1) {
            if (!s.trim()) {
                return `\`\`\`\n\`\`\`\n`;
            ***REMOVED***
            // outdent the ``` blocks, but preserve whitespace otherwise.
            const split = s.split('\n');
            // throw out the \n at the start and end
            split.pop();
            split.shift();
            const si = split.reduce((shortest, l) => {
                /* c8 ignore next */
                const ind = l.match(/^\s*/)?.[0] ?? '';
                if (ind.length)
                    return Math.min(ind.length, shortest);
                else
                    return shortest;
            ***REMOVED***, Infinity);
            /* c8 ignore next */
            const i = isFinite(si) ? si : 0;
            return ('\n```\n' +
                split.map(s => `\u200b${s.substring(i)***REMOVED***`).join('\n') +
                '\n```\n');
        ***REMOVED***
        return (s
            // remove single line breaks, except for lists
            .replace(/([^\n])\n[ \t]*([^\n])/g, (_, $1, $2) => !/^[-*]/.test($2) ? `${$1***REMOVED*** ${$2***REMOVED***` : `${$1***REMOVED***\n${$2***REMOVED***`)
            // normalize mid-line whitespace
            .replace(/([^\n])[ \t]+([^\n])/g, '$1 $2')
            // two line breaks are enough
            .replace(/\n{3,***REMOVED***/g, '\n\n')
            // remove any spaces at the start of a line
            .replace(/\n[ \t]+/g, '\n')
            .trim());
    ***REMOVED***)
        .join('\n');
***REMOVED***;
// normalize for markdown printing, remove leading spaces on lines
const normalizeMarkdown = (s, pre = false) => {
    const n = normalize(s, pre).replace(/\\/g, '\\\\');
    return pre ?
        `\`\`\`\n${n.replace(/\u200b/g, '')***REMOVED***\n\`\`\``
        : n.replace(/\n +/g, '\n').trim();
***REMOVED***;
const normalizeOneLine = (s, pre = false) => {
    const n = normalize(s, pre)
        .replace(/[\s\u200b]+/g, ' ')
        .trim();
    return pre ? `\`${n***REMOVED***\`` : n;
***REMOVED***;
/**
 * Main entry point. Create and return a {@link Jack***REMOVED*** object.
 */
const jack = (options = {***REMOVED***) => new Jack(options);
exports.jack = jack;
