// tapable/lib/HookCodeFactory.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

class HookCodeFactory {
	constructor(config) {
		this.config = config;
		this.options = undefined;
		this._args = undefined;
	***REMOVED***

	create(options) {
		this.init(options);
		let fn;
		switch (this.options.type) {
			case "sync":
				fn = new Function(
					this.args(),
					'"use strict";\n' +
						this.header() +
						this.contentWithInterceptors({
							onError: err => `throw ${err***REMOVED***;\n`,
							onResult: result => `return ${result***REMOVED***;\n`,
							resultReturns: true,
							onDone: () => "",
							rethrowIfPossible: true
						***REMOVED***)
				);
				break;
			case "async":
				fn = new Function(
					this.args({
						after: "_callback"
					***REMOVED***),
					'"use strict";\n' +
						this.header() +
						this.contentWithInterceptors({
							onError: err => `_callback(${err***REMOVED***);\n`,
							onResult: result => `_callback(null, ${result***REMOVED***);\n`,
							onDone: () => "_callback();\n"
						***REMOVED***)
				);
				break;
			case "promise":
				let errorHelperUsed = false;
				const content = this.contentWithInterceptors({
					onError: err => {
						errorHelperUsed = true;
						return `_error(${err***REMOVED***);\n`;
					***REMOVED***,
					onResult: result => `_resolve(${result***REMOVED***);\n`,
					onDone: () => "_resolve();\n"
				***REMOVED***);
				let code = "";
				code += '"use strict";\n';
				code += this.header();
				code += "return new Promise((function(_resolve, _reject) {\n";
				if (errorHelperUsed) {
					code += "var _sync = true;\n";
					code += "function _error(_err) {\n";
					code += "if(_sync)\n";
					code +=
						"_resolve(Promise.resolve().then((function() { throw _err; ***REMOVED***)));\n";
					code += "else\n";
					code += "_reject(_err);\n";
					code += "***REMOVED***;\n";
				***REMOVED***
				code += content;
				if (errorHelperUsed) {
					code += "_sync = false;\n";
				***REMOVED***
				code += "***REMOVED***));\n";
				fn = new Function(this.args(), code);
				break;
		***REMOVED***
		this.deinit();
		return fn;
	***REMOVED***

	setup(instance, options) {
		instance._x = options.taps.map(t => t.fn);
	***REMOVED***

	/**
	 * @param {{ type: "sync" | "promise" | "async", taps: Array<Tap>, interceptors: Array<Interceptor> ***REMOVED******REMOVED*** options
	 */
	init(options) {
		this.options = options;
		this._args = options.args.slice();
	***REMOVED***

	deinit() {
		this.options = undefined;
		this._args = undefined;
	***REMOVED***

	contentWithInterceptors(options) {
		if (this.options.interceptors.length > 0) {
			const onError = options.onError;
			const onResult = options.onResult;
			const onDone = options.onDone;
			let code = "";
			for (let i = 0; i < this.options.interceptors.length; i++) {
				const interceptor = this.options.interceptors[i];
				if (interceptor.call) {
					code += `${this.getInterceptor(i)***REMOVED***.call(${this.args({
						before: interceptor.context ? "_context" : undefined
					***REMOVED***)***REMOVED***);\n`;
				***REMOVED***
			***REMOVED***
			code += this.content(
				Object.assign(options, {
					onError:
						onError &&
						(err => {
							let code = "";
							for (let i = 0; i < this.options.interceptors.length; i++) {
								const interceptor = this.options.interceptors[i];
								if (interceptor.error) {
									code += `${this.getInterceptor(i)***REMOVED***.error(${err***REMOVED***);\n`;
								***REMOVED***
							***REMOVED***
							code += onError(err);
							return code;
						***REMOVED***),
					onResult:
						onResult &&
						(result => {
							let code = "";
							for (let i = 0; i < this.options.interceptors.length; i++) {
								const interceptor = this.options.interceptors[i];
								if (interceptor.result) {
									code += `${this.getInterceptor(i)***REMOVED***.result(${result***REMOVED***);\n`;
								***REMOVED***
							***REMOVED***
							code += onResult(result);
							return code;
						***REMOVED***),
					onDone:
						onDone &&
						(() => {
							let code = "";
							for (let i = 0; i < this.options.interceptors.length; i++) {
								const interceptor = this.options.interceptors[i];
								if (interceptor.done) {
									code += `${this.getInterceptor(i)***REMOVED***.done();\n`;
								***REMOVED***
							***REMOVED***
							code += onDone();
							return code;
						***REMOVED***)
				***REMOVED***)
			);
			return code;
		***REMOVED*** else {
			return this.content(options);
		***REMOVED***
	***REMOVED***

	header() {
		let code = "";
		if (this.needContext()) {
			code += "var _context = {***REMOVED***;\n";
		***REMOVED*** else {
			code += "var _context;\n";
		***REMOVED***
		code += "var _x = this._x;\n";
		if (this.options.interceptors.length > 0) {
			code += "var _taps = this.taps;\n";
			code += "var _interceptors = this.interceptors;\n";
		***REMOVED***
		return code;
	***REMOVED***

	needContext() {
		for (const tap of this.options.taps) if (tap.context) return true;
		return false;
	***REMOVED***

	callTap(tapIndex, { onError, onResult, onDone, rethrowIfPossible ***REMOVED***) {
		let code = "";
		let hasTapCached = false;
		for (let i = 0; i < this.options.interceptors.length; i++) {
			const interceptor = this.options.interceptors[i];
			if (interceptor.tap) {
				if (!hasTapCached) {
					code += `var _tap${tapIndex***REMOVED*** = ${this.getTap(tapIndex)***REMOVED***;\n`;
					hasTapCached = true;
				***REMOVED***
				code += `${this.getInterceptor(i)***REMOVED***.tap(${
					interceptor.context ? "_context, " : ""
				***REMOVED***_tap${tapIndex***REMOVED***);\n`;
			***REMOVED***
		***REMOVED***
		code += `var _fn${tapIndex***REMOVED*** = ${this.getTapFn(tapIndex)***REMOVED***;\n`;
		const tap = this.options.taps[tapIndex];
		switch (tap.type) {
			case "sync":
				if (!rethrowIfPossible) {
					code += `var _hasError${tapIndex***REMOVED*** = false;\n`;
					code += "try {\n";
				***REMOVED***
				if (onResult) {
					code += `var _result${tapIndex***REMOVED*** = _fn${tapIndex***REMOVED***(${this.args({
						before: tap.context ? "_context" : undefined
					***REMOVED***)***REMOVED***);\n`;
				***REMOVED*** else {
					code += `_fn${tapIndex***REMOVED***(${this.args({
						before: tap.context ? "_context" : undefined
					***REMOVED***)***REMOVED***);\n`;
				***REMOVED***
				if (!rethrowIfPossible) {
					code += "***REMOVED*** catch(_err) {\n";
					code += `_hasError${tapIndex***REMOVED*** = true;\n`;
					code += onError("_err");
					code += "***REMOVED***\n";
					code += `if(!_hasError${tapIndex***REMOVED***) {\n`;
				***REMOVED***
				if (onResult) {
					code += onResult(`_result${tapIndex***REMOVED***`);
				***REMOVED***
				if (onDone) {
					code += onDone();
				***REMOVED***
				if (!rethrowIfPossible) {
					code += "***REMOVED***\n";
				***REMOVED***
				break;
			case "async":
				let cbCode = "";
				if (onResult)
					cbCode += `(function(_err${tapIndex***REMOVED***, _result${tapIndex***REMOVED***) {\n`;
				else cbCode += `(function(_err${tapIndex***REMOVED***) {\n`;
				cbCode += `if(_err${tapIndex***REMOVED***) {\n`;
				cbCode += onError(`_err${tapIndex***REMOVED***`);
				cbCode += "***REMOVED*** else {\n";
				if (onResult) {
					cbCode += onResult(`_result${tapIndex***REMOVED***`);
				***REMOVED***
				if (onDone) {
					cbCode += onDone();
				***REMOVED***
				cbCode += "***REMOVED***\n";
				cbCode += "***REMOVED***)";
				code += `_fn${tapIndex***REMOVED***(${this.args({
					before: tap.context ? "_context" : undefined,
					after: cbCode
				***REMOVED***)***REMOVED***);\n`;
				break;
			case "promise":
				code += `var _hasResult${tapIndex***REMOVED*** = false;\n`;
				code += `var _promise${tapIndex***REMOVED*** = _fn${tapIndex***REMOVED***(${this.args({
					before: tap.context ? "_context" : undefined
				***REMOVED***)***REMOVED***);\n`;
				code += `if (!_promise${tapIndex***REMOVED*** || !_promise${tapIndex***REMOVED***.then)\n`;
				code += `  throw new Error('Tap function (tapPromise) did not return promise (returned ' + _promise${tapIndex***REMOVED*** + ')');\n`;
				code += `_promise${tapIndex***REMOVED***.then((function(_result${tapIndex***REMOVED***) {\n`;
				code += `_hasResult${tapIndex***REMOVED*** = true;\n`;
				if (onResult) {
					code += onResult(`_result${tapIndex***REMOVED***`);
				***REMOVED***
				if (onDone) {
					code += onDone();
				***REMOVED***
				code += `***REMOVED***), function(_err${tapIndex***REMOVED***) {\n`;
				code += `if(_hasResult${tapIndex***REMOVED***) throw _err${tapIndex***REMOVED***;\n`;
				code += onError(`_err${tapIndex***REMOVED***`);
				code += "***REMOVED***);\n";
				break;
		***REMOVED***
		return code;
	***REMOVED***

	callTapsSeries({
		onError,
		onResult,
		resultReturns,
		onDone,
		doneReturns,
		rethrowIfPossible
	***REMOVED***) {
		if (this.options.taps.length === 0) return onDone();
		const firstAsync = this.options.taps.findIndex(t => t.type !== "sync");
		const somethingReturns = resultReturns || doneReturns;
		let code = "";
		let current = onDone;
		let unrollCounter = 0;
		for (let j = this.options.taps.length - 1; j >= 0; j--) {
			const i = j;
			const unroll =
				current !== onDone &&
				(this.options.taps[i].type !== "sync" || unrollCounter++ > 20);
			if (unroll) {
				unrollCounter = 0;
				code += `function _next${i***REMOVED***() {\n`;
				code += current();
				code += `***REMOVED***\n`;
				current = () => `${somethingReturns ? "return " : ""***REMOVED***_next${i***REMOVED***();\n`;
			***REMOVED***
			const done = current;
			const doneBreak = skipDone => {
				if (skipDone) return "";
				return onDone();
			***REMOVED***;
			const content = this.callTap(i, {
				onError: error => onError(i, error, done, doneBreak),
				onResult:
					onResult &&
					(result => {
						return onResult(i, result, done, doneBreak);
					***REMOVED***),
				onDone: !onResult && done,
				rethrowIfPossible:
					rethrowIfPossible && (firstAsync < 0 || i < firstAsync)
			***REMOVED***);
			current = () => content;
		***REMOVED***
		code += current();
		return code;
	***REMOVED***

	callTapsLooping({ onError, onDone, rethrowIfPossible ***REMOVED***) {
		if (this.options.taps.length === 0) return onDone();
		const syncOnly = this.options.taps.every(t => t.type === "sync");
		let code = "";
		if (!syncOnly) {
			code += "var _looper = (function() {\n";
			code += "var _loopAsync = false;\n";
		***REMOVED***
		code += "var _loop;\n";
		code += "do {\n";
		code += "_loop = false;\n";
		for (let i = 0; i < this.options.interceptors.length; i++) {
			const interceptor = this.options.interceptors[i];
			if (interceptor.loop) {
				code += `${this.getInterceptor(i)***REMOVED***.loop(${this.args({
					before: interceptor.context ? "_context" : undefined
				***REMOVED***)***REMOVED***);\n`;
			***REMOVED***
		***REMOVED***
		code += this.callTapsSeries({
			onError,
			onResult: (i, result, next, doneBreak) => {
				let code = "";
				code += `if(${result***REMOVED*** !== undefined) {\n`;
				code += "_loop = true;\n";
				if (!syncOnly) code += "if(_loopAsync) _looper();\n";
				code += doneBreak(true);
				code += `***REMOVED*** else {\n`;
				code += next();
				code += `***REMOVED***\n`;
				return code;
			***REMOVED***,
			onDone:
				onDone &&
				(() => {
					let code = "";
					code += "if(!_loop) {\n";
					code += onDone();
					code += "***REMOVED***\n";
					return code;
				***REMOVED***),
			rethrowIfPossible: rethrowIfPossible && syncOnly
		***REMOVED***);
		code += "***REMOVED*** while(_loop);\n";
		if (!syncOnly) {
			code += "_loopAsync = true;\n";
			code += "***REMOVED***);\n";
			code += "_looper();\n";
		***REMOVED***
		return code;
	***REMOVED***

	callTapsParallel({
		onError,
		onResult,
		onDone,
		rethrowIfPossible,
		onTap = (i, run) => run()
	***REMOVED***) {
		if (this.options.taps.length <= 1) {
			return this.callTapsSeries({
				onError,
				onResult,
				onDone,
				rethrowIfPossible
			***REMOVED***);
		***REMOVED***
		let code = "";
		code += "do {\n";
		code += `var _counter = ${this.options.taps.length***REMOVED***;\n`;
		if (onDone) {
			code += "var _done = (function() {\n";
			code += onDone();
			code += "***REMOVED***);\n";
		***REMOVED***
		for (let i = 0; i < this.options.taps.length; i++) {
			const done = () => {
				if (onDone) return "if(--_counter === 0) _done();\n";
				else return "--_counter;";
			***REMOVED***;
			const doneBreak = skipDone => {
				if (skipDone || !onDone) return "_counter = 0;\n";
				else return "_counter = 0;\n_done();\n";
			***REMOVED***;
			code += "if(_counter <= 0) break;\n";
			code += onTap(
				i,
				() =>
					this.callTap(i, {
						onError: error => {
							let code = "";
							code += "if(_counter > 0) {\n";
							code += onError(i, error, done, doneBreak);
							code += "***REMOVED***\n";
							return code;
						***REMOVED***,
						onResult:
							onResult &&
							(result => {
								let code = "";
								code += "if(_counter > 0) {\n";
								code += onResult(i, result, done, doneBreak);
								code += "***REMOVED***\n";
								return code;
							***REMOVED***),
						onDone:
							!onResult &&
							(() => {
								return done();
							***REMOVED***),
						rethrowIfPossible
					***REMOVED***),
				done,
				doneBreak
			);
		***REMOVED***
		code += "***REMOVED*** while(false);\n";
		return code;
	***REMOVED***

	args({ before, after ***REMOVED*** = {***REMOVED***) {
		let allArgs = this._args;
		if (before) allArgs = [before].concat(allArgs);
		if (after) allArgs = allArgs.concat(after);
		if (allArgs.length === 0) {
			return "";
		***REMOVED*** else {
			return allArgs.join(", ");
		***REMOVED***
	***REMOVED***

	getTapFn(idx) {
		return `_x[${idx***REMOVED***]`;
	***REMOVED***

	getTap(idx) {
		return `_taps[${idx***REMOVED***]`;
	***REMOVED***

	getInterceptor(idx) {
		return `_interceptors[${idx***REMOVED***]`;
	***REMOVED***
***REMOVED***

module.exports = HookCodeFactory;
