// tapable/lib/Hook.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

const util = require("util");

const deprecateContext = util.deprecate(() => {***REMOVED***,
"Hook.context is deprecated and will be removed");

const CALL_DELEGATE = function(...args) {
	this.call = this._createCall("sync");
	return this.call(...args);
***REMOVED***;
const CALL_ASYNC_DELEGATE = function(...args) {
	this.callAsync = this._createCall("async");
	return this.callAsync(...args);
***REMOVED***;
const PROMISE_DELEGATE = function(...args) {
	this.promise = this._createCall("promise");
	return this.promise(...args);
***REMOVED***;

class Hook {
	constructor(args = [], name = undefined) {
		this._args = args;
		this.name = name;
		this.taps = [];
		this.interceptors = [];
		this._call = CALL_DELEGATE;
		this.call = CALL_DELEGATE;
		this._callAsync = CALL_ASYNC_DELEGATE;
		this.callAsync = CALL_ASYNC_DELEGATE;
		this._promise = PROMISE_DELEGATE;
		this.promise = PROMISE_DELEGATE;
		this._x = undefined;

		this.compile = this.compile;
		this.tap = this.tap;
		this.tapAsync = this.tapAsync;
		this.tapPromise = this.tapPromise;
	***REMOVED***

	compile(options) {
		throw new Error("Abstract: should be overridden");
	***REMOVED***

	_createCall(type) {
		return this.compile({
			taps: this.taps,
			interceptors: this.interceptors,
			args: this._args,
			type: type
		***REMOVED***);
	***REMOVED***

	_tap(type, options, fn) {
		if (typeof options === "string") {
			options = {
				name: options.trim()
			***REMOVED***;
		***REMOVED*** else if (typeof options !== "object" || options === null) {
			throw new Error("Invalid tap options");
		***REMOVED***
		if (typeof options.name !== "string" || options.name === "") {
			throw new Error("Missing name for tap");
		***REMOVED***
		if (typeof options.context !== "undefined") {
			deprecateContext();
		***REMOVED***
		options = Object.assign({ type, fn ***REMOVED***, options);
		options = this._runRegisterInterceptors(options);
		this._insert(options);
	***REMOVED***

	tap(options, fn) {
		this._tap("sync", options, fn);
	***REMOVED***

	tapAsync(options, fn) {
		this._tap("async", options, fn);
	***REMOVED***

	tapPromise(options, fn) {
		this._tap("promise", options, fn);
	***REMOVED***

	_runRegisterInterceptors(options) {
		for (const interceptor of this.interceptors) {
			if (interceptor.register) {
				const newOptions = interceptor.register(options);
				if (newOptions !== undefined) {
					options = newOptions;
				***REMOVED***
			***REMOVED***
		***REMOVED***
		return options;
	***REMOVED***

	withOptions(options) {
		const mergeOptions = opt =>
			Object.assign({***REMOVED***, options, typeof opt === "string" ? { name: opt ***REMOVED*** : opt);

		return {
			name: this.name,
			tap: (opt, fn) => this.tap(mergeOptions(opt), fn),
			tapAsync: (opt, fn) => this.tapAsync(mergeOptions(opt), fn),
			tapPromise: (opt, fn) => this.tapPromise(mergeOptions(opt), fn),
			intercept: interceptor => this.intercept(interceptor),
			isUsed: () => this.isUsed(),
			withOptions: opt => this.withOptions(mergeOptions(opt))
		***REMOVED***;
	***REMOVED***

	isUsed() {
		return this.taps.length > 0 || this.interceptors.length > 0;
	***REMOVED***

	intercept(interceptor) {
		this._resetCompilation();
		this.interceptors.push(Object.assign({***REMOVED***, interceptor));
		if (interceptor.register) {
			for (let i = 0; i < this.taps.length; i++) {
				this.taps[i] = interceptor.register(this.taps[i]);
			***REMOVED***
		***REMOVED***
	***REMOVED***

	_resetCompilation() {
		this.call = this._call;
		this.callAsync = this._callAsync;
		this.promise = this._promise;
	***REMOVED***

	_insert(item) {
		this._resetCompilation();
		let before;
		if (typeof item.before === "string") {
			before = new Set([item.before]);
		***REMOVED*** else if (Array.isArray(item.before)) {
			before = new Set(item.before);
		***REMOVED***
		let stage = 0;
		if (typeof item.stage === "number") {
			stage = item.stage;
		***REMOVED***
		let i = this.taps.length;
		while (i > 0) {
			i--;
			const x = this.taps[i];
			this.taps[i + 1] = x;
			const xStage = x.stage || 0;
			if (before) {
				if (before.has(x.name)) {
					before.delete(x.name);
					continue;
				***REMOVED***
				if (before.size > 0) {
					continue;
				***REMOVED***
			***REMOVED***
			if (xStage > stage) {
				continue;
			***REMOVED***
			i++;
			break;
		***REMOVED***
		this.taps[i] = item;
	***REMOVED***
***REMOVED***

Object.setPrototypeOf(Hook.prototype, null);

module.exports = Hook;
