// tapable/lib/MultiHook.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

const Hook = require("./Hook");

class MultiHook {
	constructor(hooks, name = undefined) {
		this.hooks = hooks;
		this.name = name;
	***REMOVED***

	tap(options, fn) {
		for (const hook of this.hooks) {
			hook.tap(options, fn);
		***REMOVED***
	***REMOVED***

	tapAsync(options, fn) {
		for (const hook of this.hooks) {
			hook.tapAsync(options, fn);
		***REMOVED***
	***REMOVED***

	tapPromise(options, fn) {
		for (const hook of this.hooks) {
			hook.tapPromise(options, fn);
		***REMOVED***
	***REMOVED***

	isUsed() {
		for (const hook of this.hooks) {
			if (hook.isUsed()) return true;
		***REMOVED***
		return false;
	***REMOVED***

	intercept(interceptor) {
		for (const hook of this.hooks) {
			hook.intercept(interceptor);
		***REMOVED***
	***REMOVED***

	withOptions(options) {
		return new MultiHook(
			this.hooks.map(h => h.withOptions(options)),
			this.name
		);
	***REMOVED***
***REMOVED***

module.exports = MultiHook;
