// htmlparser2/lib/MultiplexHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
/**
 * Calls a specific handler function for all events that are encountered.
 */
var MultiplexHandler = /** @class */ (function () {
    /**
     * @param func The function to multiplex all events to.
     */
    function MultiplexHandler(func) {
        this.func = func;
    ***REMOVED***
    MultiplexHandler.prototype.onattribute = function (name, value, quote) {
        this.func("onattribute", name, value, quote);
    ***REMOVED***;
    MultiplexHandler.prototype.oncdatastart = function () {
        this.func("oncdatastart");
    ***REMOVED***;
    MultiplexHandler.prototype.oncdataend = function () {
        this.func("oncdataend");
    ***REMOVED***;
    MultiplexHandler.prototype.ontext = function (text) {
        this.func("ontext", text);
    ***REMOVED***;
    MultiplexHandler.prototype.onprocessinginstruction = function (name, value) {
        this.func("onprocessinginstruction", name, value);
    ***REMOVED***;
    MultiplexHandler.prototype.oncomment = function (comment) {
        this.func("oncomment", comment);
    ***REMOVED***;
    MultiplexHandler.prototype.oncommentend = function () {
        this.func("oncommentend");
    ***REMOVED***;
    MultiplexHandler.prototype.onclosetag = function (name) {
        this.func("onclosetag", name);
    ***REMOVED***;
    MultiplexHandler.prototype.onopentag = function (name, attribs) {
        this.func("onopentag", name, attribs);
    ***REMOVED***;
    MultiplexHandler.prototype.onopentagname = function (name) {
        this.func("onopentagname", name);
    ***REMOVED***;
    MultiplexHandler.prototype.onerror = function (error) {
        this.func("onerror", error);
    ***REMOVED***;
    MultiplexHandler.prototype.onend = function () {
        this.func("onend");
    ***REMOVED***;
    MultiplexHandler.prototype.onparserinit = function (parser) {
        this.func("onparserinit", parser);
    ***REMOVED***;
    MultiplexHandler.prototype.onreset = function () {
        this.func("onreset");
    ***REMOVED***;
    return MultiplexHandler;
***REMOVED***());
exports.default = MultiplexHandler;
