// jsdom/lib/jsdom/living/constraint-validation/DefaultConstraintValidation-impl.js
"use strict";

const ValidityState = require("../generated/ValidityState");
const { isDisabled ***REMOVED*** = require("../helpers/form-controls");
const { closest ***REMOVED*** = require("../helpers/traversal");
const { fireAnEvent ***REMOVED*** = require("../helpers/events");

exports.implementation = class DefaultConstraintValidationImpl {
  // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#dom-cva-willvalidate
  get willValidate() {
    return this._isCandidateForConstraintValidation();
  ***REMOVED***

  get validity() {
    if (!this._validity) {
      this._validity = ValidityState.createImpl(this._globalObject, [], {
        element: this
      ***REMOVED***);
    ***REMOVED***
    return this._validity;
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#dom-cva-checkvalidity
  checkValidity() {
    if (!this._isCandidateForConstraintValidation()) {
      return true;
    ***REMOVED***
    if (this._satisfiesConstraints()) {
      return true;
    ***REMOVED***
    fireAnEvent("invalid", this, undefined, { cancelable: true ***REMOVED***);
    return false;
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#dom-cva-setcustomvalidity
  setCustomValidity(message) {
    this._customValidityErrorMessage = message;
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#dom-cva-reportvalidity
  // Since jsdom has no user interaction, it's the same as #checkValidity
  reportValidity() {
    return this.checkValidity();
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#dom-cva-validationmessage
  get validationMessage() {
    const { validity ***REMOVED*** = this;
    if (!this._isCandidateForConstraintValidation() || this._satisfiesConstraints()) {
      return "";
    ***REMOVED***
    const isSufferingFromCustomError = validity.customError;
    if (isSufferingFromCustomError) {
      return this._customValidityErrorMessage;
    ***REMOVED***
    return "Constraints not satisfied";
  ***REMOVED***

  _isCandidateForConstraintValidation() {
    // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fe-disabled
    return !isDisabled(this) &&
      // If an element has a datalist element ancestor,
      // it is barred from constraint validation.
      closest(this, "datalist") === null &&
      !this._barredFromConstraintValidationSpecialization();
  ***REMOVED***

  _isBarredFromConstraintValidation() {
    return !this._isCandidateForConstraintValidation();
  ***REMOVED***

  _satisfiesConstraints() {
    return this.validity.valid;
  ***REMOVED***
***REMOVED***;
