// jsdom/lib/jsdom/living/constraint-validation/ValidityState-impl.js
"use strict";

exports.implementation = class ValidityStateImpl {
  constructor(globalObject, args, privateData) {
    const { element, state = {***REMOVED*** ***REMOVED*** = privateData;

    this._element = element;
    this._state = state;
  ***REMOVED***

  get badInput() {
    return this._failsConstraint("badInput");
  ***REMOVED***

  // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#suffering-from-a-custom-error
  get customError() {
    return this._element._customValidityErrorMessage !== "";
  ***REMOVED***

  get patternMismatch() {
    return this._failsConstraint("patternMismatch");
  ***REMOVED***

  get rangeOverflow() {
    return this._failsConstraint("rangeOverflow");
  ***REMOVED***

  get rangeUnderflow() {
    return this._failsConstraint("rangeUnderflow");
  ***REMOVED***

  get stepMismatch() {
    return this._failsConstraint("stepMismatch");
  ***REMOVED***

  get tooLong() {
    return this._failsConstraint("tooLong");
  ***REMOVED***

  get tooShort() {
    return this._failsConstraint("tooShort");
  ***REMOVED***

  get typeMismatch() {
    return this._failsConstraint("typeMismatch");
  ***REMOVED***

  get valueMissing() {
    return this._failsConstraint("valueMissing");
  ***REMOVED***

  _failsConstraint(method) {
    const validationMethod = this._state[method];
    if (validationMethod) {
      return validationMethod();
    ***REMOVED***

    return false;
  ***REMOVED***

  get valid() {
    return !(this.badInput || this.valueMissing || this.customError ||
            this.patternMismatch || this.rangeOverflow || this.rangeUnderflow ||
            this.stepMismatch || this.tooLong || this.tooShort || this.typeMismatch);
  ***REMOVED***
***REMOVED***;
