// bfj/src/memory.js
'use strict'

const PROPERTIES = [ 'rss', 'heapTotal', 'heapUsed', 'external' ]

let memory

module.exports = {
  initialise,
  update,
  report
***REMOVED***

function initialise () {
  memory = PROPERTIES.reduce((result, name) => {
    result[name] = {
      sum: 0,
      hwm: 0
    ***REMOVED***
    return result
  ***REMOVED***, { count: 0 ***REMOVED***)
***REMOVED***

function update () {
  const currentMemory = process.memoryUsage()
  PROPERTIES.forEach(name => updateProperty(name, currentMemory))
***REMOVED***

function updateProperty (name, currentMemory) {
  const m = memory[name]
  const c = currentMemory[name]
  m.sum += c
  if (c > m.hwm) {
    m.hwm = c
  ***REMOVED***
***REMOVED***

function report () {
  PROPERTIES.forEach(name => reportProperty(name))
***REMOVED***

function reportProperty (name) {
  const m = memory[name]
  // eslint-disable-next-line no-console
  console.log(`mean ${name***REMOVED***: ${m.sum / memory.count***REMOVED***; hwm: ${m.hwm***REMOVED***`)
***REMOVED***
