// ajv/lib/vocabularies/validation/limitItems.ts
import type {CodeKeywordDefinition, KeywordErrorDefinition***REMOVED*** from "../../types"
import type {KeywordCxt***REMOVED*** from "../../compile/validate"
import {_, str, operators***REMOVED*** from "../../compile/codegen"

const error: KeywordErrorDefinition = {
  message({keyword, schemaCode***REMOVED***) {
    const comp = keyword === "maxItems" ? "more" : "fewer"
    return str`must NOT have ${comp***REMOVED*** than ${schemaCode***REMOVED*** items`
  ***REMOVED***,
  params: ({schemaCode***REMOVED***) => _`{limit: ${schemaCode***REMOVED******REMOVED***`,
***REMOVED***

const def: CodeKeywordDefinition = {
  keyword: ["maxItems", "minItems"],
  type: "array",
  schemaType: "number",
  $data: true,
  error,
  code(cxt: KeywordCxt) {
    const {keyword, data, schemaCode***REMOVED*** = cxt
    const op = keyword === "maxItems" ? operators.GT : operators.LT
    cxt.fail$data(_`${data***REMOVED***.length ${op***REMOVED*** ${schemaCode***REMOVED***`)
  ***REMOVED***,
***REMOVED***

export default def
