// @babel/helper-define-polyfill-provider/esm/index.node.mjs
import { declare ***REMOVED*** from '@babel/helper-plugin-utils';
import _getTargets, { prettifyTargets, getInclusionReasons, isRequired ***REMOVED*** from '@babel/helper-compilation-targets';
import * as _babel from '@babel/core';
import path from 'path';
import debounce from 'lodash.debounce';
import requireResolve from 'resolve';
import { createRequire ***REMOVED*** from 'module';

const {
  types: t$1,
  template: template
***REMOVED*** = _babel.default || _babel;
function intersection(a, b) {
  const result = new Set();
  a.forEach(v => b.has(v) && result.add(v));
  return result;
***REMOVED***
function has$1(object, key) {
  return Object.prototype.hasOwnProperty.call(object, key);
***REMOVED***
function resolve$1(path, resolved = new Set()) {
  if (resolved.has(path)) return;
  resolved.add(path);
  if (path.isVariableDeclarator()) {
    if (path.get("id").isIdentifier()) {
      return resolve$1(path.get("init"), resolved);
    ***REMOVED***
  ***REMOVED*** else if (path.isReferencedIdentifier()) {
    const binding = path.scope.getBinding(path.node.name);
    if (!binding) return path;
    if (!binding.constant) return;
    return resolve$1(binding.path, resolved);
  ***REMOVED***
  return path;
***REMOVED***
function resolveId(path) {
  if (path.isIdentifier() && !path.scope.hasBinding(path.node.name, /* noGlobals */true)) {
    return path.node.name;
  ***REMOVED***
  const resolved = resolve$1(path);
  if (resolved != null && resolved.isIdentifier()) {
    return resolved.node.name;
  ***REMOVED***
***REMOVED***
function resolveKey(path, computed = false) {
  const {
    scope
  ***REMOVED*** = path;
  if (path.isStringLiteral()) return path.node.value;
  const isIdentifier = path.isIdentifier();
  if (isIdentifier && !(computed || path.parent.computed)) {
    return path.node.name;
  ***REMOVED***
  if (computed && path.isMemberExpression() && path.get("object").isIdentifier({
    name: "Symbol"
  ***REMOVED***) && !scope.hasBinding("Symbol", /* noGlobals */true)) {
    const sym = resolveKey(path.get("property"), path.node.computed);
    if (sym) return "Symbol." + sym;
  ***REMOVED***
  if (isIdentifier ? scope.hasBinding(path.node.name, /* noGlobals */true) : path.isPure()) {
    const {
      value
    ***REMOVED*** = path.evaluate();
    if (typeof value === "string") return value;
  ***REMOVED***
***REMOVED***
function resolveSource(obj) {
  if (obj.isMemberExpression() && obj.get("property").isIdentifier({
    name: "prototype"
  ***REMOVED***)) {
    const id = resolveId(obj.get("object"));
    if (id) {
      return {
        id,
        placement: "prototype"
      ***REMOVED***;
    ***REMOVED***
    return {
      id: null,
      placement: null
    ***REMOVED***;
  ***REMOVED***
  const id = resolveId(obj);
  if (id) {
    return {
      id,
      placement: "static"
    ***REMOVED***;
  ***REMOVED***
  const path = resolve$1(obj);
  switch (path == null ? void 0 : path.type) {
    case "RegExpLiteral":
      return {
        id: "RegExp",
        placement: "prototype"
      ***REMOVED***;
    case "FunctionExpression":
      return {
        id: "Function",
        placement: "prototype"
      ***REMOVED***;
    case "StringLiteral":
      return {
        id: "String",
        placement: "prototype"
      ***REMOVED***;
    case "NumberLiteral":
      return {
        id: "Number",
        placement: "prototype"
      ***REMOVED***;
    case "BooleanLiteral":
      return {
        id: "Boolean",
        placement: "prototype"
      ***REMOVED***;
    case "ObjectExpression":
      return {
        id: "Object",
        placement: "prototype"
      ***REMOVED***;
    case "ArrayExpression":
      return {
        id: "Array",
        placement: "prototype"
      ***REMOVED***;
  ***REMOVED***
  return {
    id: null,
    placement: null
  ***REMOVED***;
***REMOVED***
function getImportSource({
  node
***REMOVED***) {
  if (node.specifiers.length === 0) return node.source.value;
***REMOVED***
function getRequireSource({
  node
***REMOVED***) {
  if (!t$1.isExpressionStatement(node)) return;
  const {
    expression
  ***REMOVED*** = node;
  if (t$1.isCallExpression(expression) && t$1.isIdentifier(expression.callee) && expression.callee.name === "require" && expression.arguments.length === 1 && t$1.isStringLiteral(expression.arguments[0])) {
    return expression.arguments[0].value;
  ***REMOVED***
***REMOVED***
function hoist(node) {
  // @ts-expect-error
  node._blockHoist = 3;
  return node;
***REMOVED***
function createUtilsGetter(cache) {
  return path => {
    const prog = path.findParent(p => p.isProgram());
    return {
      injectGlobalImport(url, moduleName) {
        cache.storeAnonymous(prog, url, moduleName, (isScript, source) => {
          return isScript ? template.statement.ast`require(${source***REMOVED***)` : t$1.importDeclaration([], source);
        ***REMOVED***);
      ***REMOVED***,
      injectNamedImport(url, name, hint = name, moduleName) {
        return cache.storeNamed(prog, url, name, moduleName, (isScript, source, name) => {
          const id = prog.scope.generateUidIdentifier(hint);
          return {
            node: isScript ? hoist(template.statement.ast`
                  var ${id***REMOVED*** = require(${source***REMOVED***).${name***REMOVED***
                `) : t$1.importDeclaration([t$1.importSpecifier(id, name)], source),
            name: id.name
          ***REMOVED***;
        ***REMOVED***);
      ***REMOVED***,
      injectDefaultImport(url, hint = url, moduleName) {
        return cache.storeNamed(prog, url, "default", moduleName, (isScript, source) => {
          const id = prog.scope.generateUidIdentifier(hint);
          return {
            node: isScript ? hoist(template.statement.ast`var ${id***REMOVED*** = require(${source***REMOVED***)`) : t$1.importDeclaration([t$1.importDefaultSpecifier(id)], source),
            name: id.name
          ***REMOVED***;
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***;
***REMOVED***

const {
  types: t
***REMOVED*** = _babel.default || _babel;
class ImportsCachedInjector {
  constructor(resolver, getPreferredIndex) {
    this._imports = new WeakMap();
    this._anonymousImports = new WeakMap();
    this._lastImports = new WeakMap();
    this._resolver = resolver;
    this._getPreferredIndex = getPreferredIndex;
  ***REMOVED***
  storeAnonymous(programPath, url, moduleName, getVal) {
    const key = this._normalizeKey(programPath, url);
    const imports = this._ensure(this._anonymousImports, programPath, Set);
    if (imports.has(key)) return;
    const node = getVal(programPath.node.sourceType === "script", t.stringLiteral(this._resolver(url)));
    imports.add(key);
    this._injectImport(programPath, node, moduleName);
  ***REMOVED***
  storeNamed(programPath, url, name, moduleName, getVal) {
    const key = this._normalizeKey(programPath, url, name);
    const imports = this._ensure(this._imports, programPath, Map);
    if (!imports.has(key)) {
      const {
        node,
        name: id
      ***REMOVED*** = getVal(programPath.node.sourceType === "script", t.stringLiteral(this._resolver(url)), t.identifier(name));
      imports.set(key, id);
      this._injectImport(programPath, node, moduleName);
    ***REMOVED***
    return t.identifier(imports.get(key));
  ***REMOVED***
  _injectImport(programPath, node, moduleName) {
    var _this$_lastImports$ge;
    const newIndex = this._getPreferredIndex(moduleName);
    const lastImports = (_this$_lastImports$ge = this._lastImports.get(programPath)) != null ? _this$_lastImports$ge : [];
    const isPathStillValid = path => path.node &&
    // Sometimes the AST is modified and the "last import"
    // we have has been replaced
    path.parent === programPath.node && path.container === programPath.node.body;
    let last;
    if (newIndex === Infinity) {
      // Fast path: we can always just insert at the end if newIndex is `Infinity`
      if (lastImports.length > 0) {
        last = lastImports[lastImports.length - 1].path;
        if (!isPathStillValid(last)) last = undefined;
      ***REMOVED***
    ***REMOVED*** else {
      for (const [i, data] of lastImports.entries()) {
        const {
          path,
          index
        ***REMOVED*** = data;
        if (isPathStillValid(path)) {
          if (newIndex < index) {
            const [newPath] = path.insertBefore(node);
            lastImports.splice(i, 0, {
              path: newPath,
              index: newIndex
            ***REMOVED***);
            return;
          ***REMOVED***
          last = path;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
    if (last) {
      const [newPath] = last.insertAfter(node);
      lastImports.push({
        path: newPath,
        index: newIndex
      ***REMOVED***);
    ***REMOVED*** else {
      const [newPath] = programPath.unshiftContainer("body", node);
      this._lastImports.set(programPath, [{
        path: newPath,
        index: newIndex
      ***REMOVED***]);
    ***REMOVED***
  ***REMOVED***
  _ensure(map, programPath, Collection) {
    let collection = map.get(programPath);
    if (!collection) {
      collection = new Collection();
      map.set(programPath, collection);
    ***REMOVED***
    return collection;
  ***REMOVED***
  _normalizeKey(programPath, url, name = "") {
    const {
      sourceType
    ***REMOVED*** = programPath.node;

    // If we rely on the imported binding (the "name" parameter), we also need to cache
    // based on the sourceType. This is because the module transforms change the names
    // of the import variables.
    return `${name && sourceType***REMOVED***::${url***REMOVED***::${name***REMOVED***`;
  ***REMOVED***
***REMOVED***

const presetEnvSilentDebugHeader = "#__secret_key__@babel/preset-env__don't_log_debug_header_and_resolved_targets";
function stringifyTargetsMultiline(targets) {
  return JSON.stringify(prettifyTargets(targets), null, 2);
***REMOVED***

function patternToRegExp(pattern) {
  if (pattern instanceof RegExp) return pattern;
  try {
    return new RegExp(`^${pattern***REMOVED***$`);
  ***REMOVED*** catch {
    return null;
  ***REMOVED***
***REMOVED***
function buildUnusedError(label, unused) {
  if (!unused.length) return "";
  return `  - The following "${label***REMOVED***" patterns didn't match any polyfill:\n` + unused.map(original => `    ${String(original)***REMOVED***\n`).join("");
***REMOVED***
function buldDuplicatesError(duplicates) {
  if (!duplicates.size) return "";
  return `  - The following polyfills were matched both by "include" and "exclude" patterns:\n` + Array.from(duplicates, name => `    ${name***REMOVED***\n`).join("");
***REMOVED***
function validateIncludeExclude(provider, polyfills, includePatterns, excludePatterns) {
  let current;
  const filter = pattern => {
    const regexp = patternToRegExp(pattern);
    if (!regexp) return false;
    let matched = false;
    for (const polyfill of polyfills.keys()) {
      if (regexp.test(polyfill)) {
        matched = true;
        current.add(polyfill);
      ***REMOVED***
    ***REMOVED***
    return !matched;
  ***REMOVED***;

  // prettier-ignore
  const include = current = new Set();
  const unusedInclude = Array.from(includePatterns).filter(filter);

  // prettier-ignore
  const exclude = current = new Set();
  const unusedExclude = Array.from(excludePatterns).filter(filter);
  const duplicates = intersection(include, exclude);
  if (duplicates.size > 0 || unusedInclude.length > 0 || unusedExclude.length > 0) {
    throw new Error(`Error while validating the "${provider***REMOVED***" provider options:\n` + buildUnusedError("include", unusedInclude) + buildUnusedError("exclude", unusedExclude) + buldDuplicatesError(duplicates));
  ***REMOVED***
  return {
    include,
    exclude
  ***REMOVED***;
***REMOVED***
function applyMissingDependenciesDefaults(options, babelApi) {
  const {
    missingDependencies = {***REMOVED***
  ***REMOVED*** = options;
  if (missingDependencies === false) return false;
  const caller = babelApi.caller(caller => caller == null ? void 0 : caller.name);
  const {
    log = "deferred",
    inject = caller === "rollup-plugin-babel" ? "throw" : "import",
    all = false
  ***REMOVED*** = missingDependencies;
  return {
    log,
    inject,
    all
  ***REMOVED***;
***REMOVED***

function isRemoved(path) {
  if (path.removed) return true;
  if (!path.parentPath) return false;
  if (path.listKey) {
    var _path$parentPath$node;
    if (!((_path$parentPath$node = path.parentPath.node) != null && (_path$parentPath$node = _path$parentPath$node[path.listKey]) != null && _path$parentPath$node.includes(path.node))) return true;
  ***REMOVED*** else {
    if (path.parentPath.node[path.key] !== path.node) return true;
  ***REMOVED***
  return isRemoved(path.parentPath);
***REMOVED***
var usage = (callProvider => {
  function property(object, key, placement, path) {
    return callProvider({
      kind: "property",
      object,
      key,
      placement
    ***REMOVED***, path);
  ***REMOVED***
  function handleReferencedIdentifier(path) {
    const {
      node: {
        name
      ***REMOVED***,
      scope
    ***REMOVED*** = path;
    if (scope.getBindingIdentifier(name)) return;
    callProvider({
      kind: "global",
      name
    ***REMOVED***, path);
  ***REMOVED***
  function analyzeMemberExpression(path) {
    const key = resolveKey(path.get("property"), path.node.computed);
    return {
      key,
      handleAsMemberExpression: !!key && key !== "prototype"
    ***REMOVED***;
  ***REMOVED***
  return {
    // Symbol(), new Promise
    ReferencedIdentifier(path) {
      const {
        parentPath
      ***REMOVED*** = path;
      if (parentPath.isMemberExpression({
        object: path.node
      ***REMOVED***) && analyzeMemberExpression(parentPath).handleAsMemberExpression) {
        return;
      ***REMOVED***
      handleReferencedIdentifier(path);
    ***REMOVED***,
    "MemberExpression|OptionalMemberExpression"(path) {
      const {
        key,
        handleAsMemberExpression
      ***REMOVED*** = analyzeMemberExpression(path);
      if (!handleAsMemberExpression) return;
      const object = path.get("object");
      let objectIsGlobalIdentifier = object.isIdentifier();
      if (objectIsGlobalIdentifier) {
        const binding = object.scope.getBinding(object.node.name);
        if (binding) {
          if (binding.path.isImportNamespaceSpecifier()) return;
          objectIsGlobalIdentifier = false;
        ***REMOVED***
      ***REMOVED***
      const source = resolveSource(object);
      let skipObject = property(source.id, key, source.placement, path);
      skipObject || (skipObject = !objectIsGlobalIdentifier || path.shouldSkip || object.shouldSkip || isRemoved(object));
      if (!skipObject) handleReferencedIdentifier(object);
    ***REMOVED***,
    ObjectPattern(path) {
      const {
        parentPath,
        parent
      ***REMOVED*** = path;
      let obj;

      // const { keys, values ***REMOVED*** = Object
      if (parentPath.isVariableDeclarator()) {
        obj = parentPath.get("init");
        // ({ keys, values ***REMOVED*** = Object)
      ***REMOVED*** else if (parentPath.isAssignmentExpression()) {
        obj = parentPath.get("right");
        // !function ({ keys, values ***REMOVED***) {...***REMOVED*** (Object)
        // resolution does not work after properties transform :-(
      ***REMOVED*** else if (parentPath.isFunction()) {
        const grand = parentPath.parentPath;
        if (grand.isCallExpression() || grand.isNewExpression()) {
          if (grand.node.callee === parent) {
            obj = grand.get("arguments")[path.key];
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
      let id = null;
      let placement = null;
      if (obj) ({
        id,
        placement
      ***REMOVED*** = resolveSource(obj));
      for (const prop of path.get("properties")) {
        if (prop.isObjectProperty()) {
          const key = resolveKey(prop.get("key"));
          if (key) property(id, key, placement, prop);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***,
    BinaryExpression(path) {
      if (path.node.operator !== "in") return;
      const source = resolveSource(path.get("right"));
      const key = resolveKey(path.get("left"), true);
      if (!key) return;
      callProvider({
        kind: "in",
        object: source.id,
        key,
        placement: source.placement
      ***REMOVED***, path);
    ***REMOVED***
  ***REMOVED***;
***REMOVED***);

var entry = (callProvider => ({
  ImportDeclaration(path) {
    const source = getImportSource(path);
    if (!source) return;
    callProvider({
      kind: "import",
      source
    ***REMOVED***, path);
  ***REMOVED***,
  Program(path) {
    path.get("body").forEach(bodyPath => {
      const source = getRequireSource(bodyPath);
      if (!source) return;
      callProvider({
        kind: "import",
        source
      ***REMOVED***, bodyPath);
    ***REMOVED***);
  ***REMOVED***
***REMOVED***));

const nativeRequireResolve = parseFloat(process.versions.node) >= 8.9;
const require = createRequire(import /*::(_)*/.meta.url); // eslint-disable-line

function myResolve(name, basedir) {
  if (nativeRequireResolve) {
    return require.resolve(name, {
      paths: [basedir]
    ***REMOVED***).replace(/\\/g, "/");
  ***REMOVED*** else {
    return requireResolve.sync(name, {
      basedir
    ***REMOVED***).replace(/\\/g, "/");
  ***REMOVED***
***REMOVED***
function resolve(dirname, moduleName, absoluteImports) {
  if (absoluteImports === false) return moduleName;
  let basedir = dirname;
  if (typeof absoluteImports === "string") {
    basedir = path.resolve(basedir, absoluteImports);
  ***REMOVED***
  try {
    return myResolve(moduleName, basedir);
  ***REMOVED*** catch (err) {
    if (err.code !== "MODULE_NOT_FOUND") throw err;
    throw Object.assign(new Error(`Failed to resolve "${moduleName***REMOVED***" relative to "${dirname***REMOVED***"`), {
      code: "BABEL_POLYFILL_NOT_FOUND",
      polyfill: moduleName,
      dirname
    ***REMOVED***);
  ***REMOVED***
***REMOVED***
function has(basedir, name) {
  try {
    myResolve(name, basedir);
    return true;
  ***REMOVED*** catch {
    return false;
  ***REMOVED***
***REMOVED***
function logMissing(missingDeps) {
  if (missingDeps.size === 0) return;
  const deps = Array.from(missingDeps).sort().join(" ");
  console.warn("\nSome polyfills have been added but are not present in your dependencies.\n" + "Please run one of the following commands:\n" + `\tnpm install --save ${deps***REMOVED***\n` + `\tyarn add ${deps***REMOVED***\n`);
  process.exitCode = 1;
***REMOVED***
let allMissingDeps = new Set();
const laterLogMissingDependencies = debounce(() => {
  logMissing(allMissingDeps);
  allMissingDeps = new Set();
***REMOVED***, 100);
function laterLogMissing(missingDeps) {
  if (missingDeps.size === 0) return;
  missingDeps.forEach(name => allMissingDeps.add(name));
  laterLogMissingDependencies();
***REMOVED***

const PossibleGlobalObjects = new Set(["global", "globalThis", "self", "window"]);
function createMetaResolver(polyfills) {
  const {
    static: staticP,
    instance: instanceP,
    global: globalP
  ***REMOVED*** = polyfills;
  return meta => {
    if (meta.kind === "global" && globalP && has$1(globalP, meta.name)) {
      return {
        kind: "global",
        desc: globalP[meta.name],
        name: meta.name
      ***REMOVED***;
    ***REMOVED***
    if (meta.kind === "property" || meta.kind === "in") {
      const {
        placement,
        object,
        key
      ***REMOVED*** = meta;
      if (object && placement === "static") {
        if (globalP && PossibleGlobalObjects.has(object) && has$1(globalP, key)) {
          return {
            kind: "global",
            desc: globalP[key],
            name: key
          ***REMOVED***;
        ***REMOVED***
        if (staticP && has$1(staticP, object) && has$1(staticP[object], key)) {
          return {
            kind: "static",
            desc: staticP[object][key],
            name: `${object***REMOVED***$${key***REMOVED***`
          ***REMOVED***;
        ***REMOVED***
      ***REMOVED***
      if (instanceP && has$1(instanceP, key)) {
        return {
          kind: "instance",
          desc: instanceP[key],
          name: `${key***REMOVED***`
        ***REMOVED***;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;
***REMOVED***

const getTargets = _getTargets.default || _getTargets;
function resolveOptions(options, babelApi) {
  const {
    method,
    targets: targetsOption,
    ignoreBrowserslistConfig,
    configPath,
    debug,
    shouldInjectPolyfill,
    absoluteImports,
    ...providerOptions
  ***REMOVED*** = options;
  if (isEmpty(options)) {
    throw new Error(`\
This plugin requires options, for example:
    {
      "plugins": [
        ["<plugin name>", { method: "usage-pure" ***REMOVED***]
      ]
    ***REMOVED***

See more options at https://github.com/babel/babel-polyfills/blob/main/docs/usage.md`);
  ***REMOVED***
  let methodName;
  if (method === "usage-global") methodName = "usageGlobal";else if (method === "entry-global") methodName = "entryGlobal";else if (method === "usage-pure") methodName = "usagePure";else if (typeof method !== "string") {
    throw new Error(".method must be a string");
  ***REMOVED*** else {
    throw new Error(`.method must be one of "entry-global", "usage-global"` + ` or "usage-pure" (received ${JSON.stringify(method)***REMOVED***)`);
  ***REMOVED***
  if (typeof shouldInjectPolyfill === "function") {
    if (options.include || options.exclude) {
      throw new Error(`.include and .exclude are not supported when using the` + ` .shouldInjectPolyfill function.`);
    ***REMOVED***
  ***REMOVED*** else if (shouldInjectPolyfill != null) {
    throw new Error(`.shouldInjectPolyfill must be a function, or undefined` + ` (received ${JSON.stringify(shouldInjectPolyfill)***REMOVED***)`);
  ***REMOVED***
  if (absoluteImports != null && typeof absoluteImports !== "boolean" && typeof absoluteImports !== "string") {
    throw new Error(`.absoluteImports must be a boolean, a string, or undefined` + ` (received ${JSON.stringify(absoluteImports)***REMOVED***)`);
  ***REMOVED***
  let targets;
  if (
  // If any browserslist-related option is specified, fallback to the old
  // behavior of not using the targets specified in the top-level options.
  targetsOption || configPath || ignoreBrowserslistConfig) {
    const targetsObj = typeof targetsOption === "string" || Array.isArray(targetsOption) ? {
      browsers: targetsOption
    ***REMOVED*** : targetsOption;
    targets = getTargets(targetsObj, {
      ignoreBrowserslistConfig,
      configPath
    ***REMOVED***);
  ***REMOVED*** else {
    targets = babelApi.targets();
  ***REMOVED***
  return {
    method,
    methodName,
    targets,
    absoluteImports: absoluteImports != null ? absoluteImports : false,
    shouldInjectPolyfill,
    debug: !!debug,
    providerOptions: providerOptions
  ***REMOVED***;
***REMOVED***
function instantiateProvider(factory, options, missingDependencies, dirname, debugLog, babelApi) {
  const {
    method,
    methodName,
    targets,
    debug,
    shouldInjectPolyfill,
    providerOptions,
    absoluteImports
  ***REMOVED*** = resolveOptions(options, babelApi);

  // eslint-disable-next-line prefer-const
  let include, exclude;
  let polyfillsSupport;
  let polyfillsNames;
  let filterPolyfills;
  const getUtils = createUtilsGetter(new ImportsCachedInjector(moduleName => resolve(dirname, moduleName, absoluteImports), name => {
    var _polyfillsNames$get, _polyfillsNames;
    return (_polyfillsNames$get = (_polyfillsNames = polyfillsNames) == null ? void 0 : _polyfillsNames.get(name)) != null ? _polyfillsNames$get : Infinity;
  ***REMOVED***));
  const depsCache = new Map();
  const api = {
    babel: babelApi,
    getUtils,
    method: options.method,
    targets,
    createMetaResolver,
    shouldInjectPolyfill(name) {
      if (polyfillsNames === undefined) {
        throw new Error(`Internal error in the ${factory.name***REMOVED*** provider: ` + `shouldInjectPolyfill() can't be called during initialization.`);
      ***REMOVED***
      if (!polyfillsNames.has(name)) {
        console.warn(`Internal error in the ${providerName***REMOVED*** provider: ` + `unknown polyfill "${name***REMOVED***".`);
      ***REMOVED***
      if (filterPolyfills && !filterPolyfills(name)) return false;
      let shouldInject = isRequired(name, targets, {
        compatData: polyfillsSupport,
        includes: include,
        excludes: exclude
      ***REMOVED***);
      if (shouldInjectPolyfill) {
        shouldInject = shouldInjectPolyfill(name, shouldInject);
        if (typeof shouldInject !== "boolean") {
          throw new Error(`.shouldInjectPolyfill must return a boolean.`);
        ***REMOVED***
      ***REMOVED***
      return shouldInject;
    ***REMOVED***,
    debug(name) {
      var _debugLog, _debugLog$polyfillsSu;
      debugLog().found = true;
      if (!debug || !name) return;
      if (debugLog().polyfills.has(providerName)) return;
      debugLog().polyfills.add(name);
      (_debugLog$polyfillsSu = (_debugLog = debugLog()).polyfillsSupport) != null ? _debugLog$polyfillsSu : _debugLog.polyfillsSupport = polyfillsSupport;
    ***REMOVED***,
    assertDependency(name, version = "*") {
      if (missingDependencies === false) return;
      if (absoluteImports) {
        // If absoluteImports is not false, we will try resolving
        // the dependency and throw if it's not possible. We can
        // skip the check here.
        return;
      ***REMOVED***
      const dep = version === "*" ? name : `${name***REMOVED***@^${version***REMOVED***`;
      const found = missingDependencies.all ? false : mapGetOr(depsCache, `${name***REMOVED*** :: ${dirname***REMOVED***`, () => has(dirname, name));
      if (!found) {
        debugLog().missingDeps.add(dep);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;
  const provider = factory(api, providerOptions, dirname);
  const providerName = provider.name || factory.name;
  if (typeof provider[methodName] !== "function") {
    throw new Error(`The "${providerName***REMOVED***" provider doesn't support the "${method***REMOVED***" polyfilling method.`);
  ***REMOVED***
  if (Array.isArray(provider.polyfills)) {
    polyfillsNames = new Map(provider.polyfills.map((name, index) => [name, index]));
    filterPolyfills = provider.filterPolyfills;
  ***REMOVED*** else if (provider.polyfills) {
    polyfillsNames = new Map(Object.keys(provider.polyfills).map((name, index) => [name, index]));
    polyfillsSupport = provider.polyfills;
    filterPolyfills = provider.filterPolyfills;
  ***REMOVED*** else {
    polyfillsNames = new Map();
  ***REMOVED***
  ({
    include,
    exclude
  ***REMOVED*** = validateIncludeExclude(providerName, polyfillsNames, providerOptions.include || [], providerOptions.exclude || []));
  let callProvider;
  if (methodName === "usageGlobal") {
    callProvider = (payload, path) => {
      var _ref;
      const utils = getUtils(path);
      return (_ref = provider[methodName](payload, utils, path)) != null ? _ref : false;
    ***REMOVED***;
  ***REMOVED*** else {
    callProvider = (payload, path) => {
      const utils = getUtils(path);
      provider[methodName](payload, utils, path);
      return false;
    ***REMOVED***;
  ***REMOVED***
  return {
    debug,
    method,
    targets,
    provider,
    providerName,
    callProvider
  ***REMOVED***;
***REMOVED***
function definePolyfillProvider(factory) {
  return declare((babelApi, options, dirname) => {
    babelApi.assertVersion("^7.0.0 || ^8.0.0-alpha.0");
    const {
      traverse
    ***REMOVED*** = babelApi;
    let debugLog;
    const missingDependencies = applyMissingDependenciesDefaults(options, babelApi);
    const {
      debug,
      method,
      targets,
      provider,
      providerName,
      callProvider
    ***REMOVED*** = instantiateProvider(factory, options, missingDependencies, dirname, () => debugLog, babelApi);
    const createVisitor = method === "entry-global" ? entry : usage;
    const visitor = provider.visitor ? traverse.visitors.merge([createVisitor(callProvider), provider.visitor]) : createVisitor(callProvider);
    if (debug && debug !== presetEnvSilentDebugHeader) {
      console.log(`${providerName***REMOVED***: \`DEBUG\` option`);
      console.log(`\nUsing targets: ${stringifyTargetsMultiline(targets)***REMOVED***`);
      console.log(`\nUsing polyfills with \`${method***REMOVED***\` method:`);
    ***REMOVED***
    const {
      runtimeName
    ***REMOVED*** = provider;
    return {
      name: "inject-polyfills",
      visitor,
      pre(file) {
        var _provider$pre;
        if (runtimeName) {
          if (file.get("runtimeHelpersModuleName") && file.get("runtimeHelpersModuleName") !== runtimeName) {
            console.warn(`Two different polyfill providers` + ` (${file.get("runtimeHelpersModuleProvider")***REMOVED***` + ` and ${providerName***REMOVED***) are trying to define two` + ` conflicting @babel/runtime alternatives:` + ` ${file.get("runtimeHelpersModuleName")***REMOVED*** and ${runtimeName***REMOVED***.` + ` The second one will be ignored.`);
          ***REMOVED*** else {
            file.set("runtimeHelpersModuleName", runtimeName);
            file.set("runtimeHelpersModuleProvider", providerName);
          ***REMOVED***
        ***REMOVED***
        debugLog = {
          polyfills: new Set(),
          polyfillsSupport: undefined,
          found: false,
          providers: new Set(),
          missingDeps: new Set()
        ***REMOVED***;
        (_provider$pre = provider.pre) == null ? void 0 : _provider$pre.apply(this, arguments);
      ***REMOVED***,
      post() {
        var _provider$post;
        (_provider$post = provider.post) == null ? void 0 : _provider$post.apply(this, arguments);
        if (missingDependencies !== false) {
          if (missingDependencies.log === "per-file") {
            logMissing(debugLog.missingDeps);
          ***REMOVED*** else {
            laterLogMissing(debugLog.missingDeps);
          ***REMOVED***
        ***REMOVED***
        if (!debug) return;
        if (this.filename) console.log(`\n[${this.filename***REMOVED***]`);
        if (debugLog.polyfills.size === 0) {
          console.log(method === "entry-global" ? debugLog.found ? `Based on your targets, the ${providerName***REMOVED*** polyfill did not add any polyfill.` : `The entry point for the ${providerName***REMOVED*** polyfill has not been found.` : `Based on your code and targets, the ${providerName***REMOVED*** polyfill did not add any polyfill.`);
          return;
        ***REMOVED***
        if (method === "entry-global") {
          console.log(`The ${providerName***REMOVED*** polyfill entry has been replaced with ` + `the following polyfills:`);
        ***REMOVED*** else {
          console.log(`The ${providerName***REMOVED*** polyfill added the following polyfills:`);
        ***REMOVED***
        for (const name of debugLog.polyfills) {
          var _debugLog$polyfillsSu2;
          if ((_debugLog$polyfillsSu2 = debugLog.polyfillsSupport) != null && _debugLog$polyfillsSu2[name]) {
            const filteredTargets = getInclusionReasons(name, targets, debugLog.polyfillsSupport);
            const formattedTargets = JSON.stringify(filteredTargets).replace(/,/g, ", ").replace(/^\{"/, '{ "').replace(/"\***REMOVED***$/, '" ***REMOVED***');
            console.log(`  ${name***REMOVED*** ${formattedTargets***REMOVED***`);
          ***REMOVED*** else {
            console.log(`  ${name***REMOVED***`);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***);
***REMOVED***
function mapGetOr(map, key, getDefault) {
  let val = map.get(key);
  if (val === undefined) {
    val = getDefault();
    map.set(key, val);
  ***REMOVED***
  return val;
***REMOVED***
function isEmpty(obj) {
  return Object.keys(obj).length === 0;
***REMOVED***

export default definePolyfillProvider;
