// jsdom/lib/jsdom/living/xhr/FormData-impl.js
"use strict";
const idlUtils = require("../generated/utils");
const { closest ***REMOVED*** = require("../helpers/traversal");
const { isDisabled, isSubmittable, isButton ***REMOVED*** = require("../helpers/form-controls");
const Blob = require("../generated/Blob.js");
const File = require("../generated/File.js");
const conversions = require("webidl-conversions");

exports.implementation = class FormDataImpl {
  constructor(globalObject, args) {
    this._globalObject = globalObject;
    this._entries = [];

    if (args[0] !== undefined) {
      this._entries = constructTheEntryList(args[0]);
    ***REMOVED***
  ***REMOVED***

  append(name, value, filename) {
    const entry = createAnEntry(name, value, filename);
    this._entries.push(entry);
  ***REMOVED***

  delete(name) {
    this._entries = this._entries.filter(entry => entry.name !== name);
  ***REMOVED***

  get(name) {
    const foundEntry = this._entries.find(entry => entry.name === name);
    return foundEntry !== undefined ? idlUtils.tryWrapperForImpl(foundEntry.value) : null;
  ***REMOVED***

  getAll(name) {
    return this._entries.filter(entry => entry.name === name).map(entry => idlUtils.tryWrapperForImpl(entry.value));
  ***REMOVED***

  has(name) {
    return this._entries.findIndex(entry => entry.name === name) !== -1;
  ***REMOVED***

  set(name, value, filename) {
    const entry = createAnEntry(name, value, filename);

    const foundIndex = this._entries.findIndex(e => e.name === name);
    if (foundIndex !== -1) {
      this._entries[foundIndex] = entry;
      this._entries = this._entries.filter((e, i) => e.name !== name || i === foundIndex);
    ***REMOVED*** else {
      this._entries.push(entry);
    ***REMOVED***
  ***REMOVED***

  * [Symbol.iterator]() {
    for (const entry of this._entries) {
      yield [entry.name, idlUtils.tryWrapperForImpl(entry.value)];
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

function createAnEntry(name, value, filename) {
  const entry = { name ***REMOVED***;

  // https://github.com/whatwg/xhr/issues/75

  if (Blob.isImpl(value) && !File.isImpl(value)) {
    const oldValue = value;
    value = File.createImpl(value._globalObject, [
      [],
      "blob",
      { type: oldValue.type ***REMOVED***
    ]);
    // "representing the same bytes"
    value._buffer = oldValue._buffer;
  ***REMOVED***

  if (File.isImpl(value) && filename !== undefined) {
    const oldValue = value;
    value = File.createImpl(value._globalObject, [
      [],
      filename,
      // spec makes no mention of `lastModified`; assume it is inherited
      // (Chrome's behavior)
      { type: oldValue.type, lastModified: oldValue.lastModified ***REMOVED***
    ]);
    // "representing the same bytes"
    value._buffer = oldValue._buffer;
  ***REMOVED***

  entry.value = value;

  return entry;
***REMOVED***

function constructTheEntryList(form, submitter) {
  // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#constructing-the-form-data-set
  // TODO: handle encoding
  // TODO: handling "constructing entry list"

  const controls = form.elements.filter(isSubmittable); // submittable is a subset of listed
  const entryList = [];

  for (const field of controls) {
    if (closest(field, "datalist") !== null) {
      continue;
    ***REMOVED***
    if (isDisabled(field)) {
      continue;
    ***REMOVED***
    if (isButton(field) && field !== submitter) {
      continue;
    ***REMOVED***
    if (field.type === "checkbox" && field._checkedness === false) {
      continue;
    ***REMOVED***
    if (field.type === "radio" && field._checkedness === false) {
      continue;
    ***REMOVED***
    if (field.localName === "object") { // in jsdom, no objects are "using a plugin"
      continue;
    ***REMOVED***

    // TODO: Handle <input type="image">
    // TODO: handle form-associated custom elements.

    const name = field.getAttributeNS(null, "name");
    if (name === null || name === "") {
      continue;
    ***REMOVED***

    if (field.localName === "select") {
      for (const option of field.options) {
        if (option._selectedness === true && !isDisabled(field)) {
          appendAnEntry(entryList, name, option._getValue());
        ***REMOVED***
      ***REMOVED***
    ***REMOVED*** else if (field.localName === "input" && (field.type === "checkbox" || field.type === "radio")) {
      const value = field.hasAttributeNS(null, "value") ? field.getAttributeNS(null, "value") : "on";
      appendAnEntry(entryList, name, value);
    ***REMOVED*** else if (field.type === "file") {
      if (field.files.length === 0) {
        const value = File.createImpl(form._globalObject, [[], "", { type: "application/octet-stream" ***REMOVED***]);
        appendAnEntry(entryList, name, value);
      ***REMOVED*** else {
        for (let i = 0; i < field.files.length; ++i) {
          appendAnEntry(entryList, name, field.files.item(i));
        ***REMOVED***
      ***REMOVED***
    ***REMOVED*** else {
      appendAnEntry(entryList, name, field._getValue());
    ***REMOVED***

    const dirname = field.getAttributeNS(null, "dirname");
    if (dirname !== null && dirname !== "") {
      const dir = "ltr"; // jsdom does not (yet?) implement actual directionality
      appendAnEntry(entryList, dirname, dir);
    ***REMOVED***
  ***REMOVED***

  // TODO: formdata event

  return entryList;
***REMOVED***

function appendAnEntry(entryList, name, value) {
  name = conversions.USVString(name);
  if (!File.isImpl(value)) {
    value = conversions.USVString(value);
  ***REMOVED***
  const entry = createAnEntry(name, value);
  entryList.push(entry);
***REMOVED***
