// istanbul-lib-report/lib/context.js
'use strict';
/*
 Copyright 2012-2015, Yahoo Inc.
 Copyrights licensed under the New BSD License. See the accompanying LICENSE file for terms.
 */
const fs = require('fs');
const FileWriter = require('./file-writer');
const XMLWriter = require('./xml-writer');
const tree = require('./tree');
const watermarks = require('./watermarks');
const SummarizerFactory = require('./summarizer-factory');

function defaultSourceLookup(path) {
    try {
        return fs.readFileSync(path, 'utf8');
    ***REMOVED*** catch (ex) {
        throw new Error(`Unable to lookup source: ${path***REMOVED*** (${ex.message***REMOVED***)`);
    ***REMOVED***
***REMOVED***

function normalizeWatermarks(specified = {***REMOVED***) {
    Object.entries(watermarks.getDefault()).forEach(([k, value]) => {
        const specValue = specified[k];
        if (!Array.isArray(specValue) || specValue.length !== 2) {
            specified[k] = value;
        ***REMOVED***
    ***REMOVED***);

    return specified;
***REMOVED***

/**
 * A reporting context that is passed to report implementations
 * @param {Object***REMOVED*** [opts=null] opts options
 * @param {String***REMOVED*** [opts.dir='coverage'] opts.dir the reporting directory
 * @param {Object***REMOVED*** [opts.watermarks=null] opts.watermarks watermarks for
 *  statements, lines, branches and functions
 * @param {Function***REMOVED*** [opts.sourceFinder=fsLookup] opts.sourceFinder a
 *  function that returns source code given a file path. Defaults to
 *  filesystem lookups based on path.
 * @constructor
 */
class Context {
    constructor(opts) {
        this.dir = opts.dir || 'coverage';
        this.watermarks = normalizeWatermarks(opts.watermarks);
        this.sourceFinder = opts.sourceFinder || defaultSourceLookup;
        this._summarizerFactory = new SummarizerFactory(
            opts.coverageMap,
            opts.defaultSummarizer
        );
        this.data = {***REMOVED***;
    ***REMOVED***

    /**
     * returns a FileWriter implementation for reporting use. Also available
     * as the `writer` property on the context.
     * @returns {Writer***REMOVED***
     */
    getWriter() {
        return this.writer;
    ***REMOVED***

    /**
     * returns the source code for the specified file path or throws if
     * the source could not be found.
     * @param {String***REMOVED*** filePath the file path as found in a file coverage object
     * @returns {String***REMOVED*** the source code
     */
    getSource(filePath) {
        return this.sourceFinder(filePath);
    ***REMOVED***

    /**
     * returns the coverage class given a coverage
     * types and a percentage value.
     * @param {String***REMOVED*** type - the coverage type, one of `statements`, `functions`,
     *  `branches`, or `lines`
     * @param {Number***REMOVED*** value - the percentage value
     * @returns {String***REMOVED*** one of `high`, `medium` or `low`
     */
    classForPercent(type, value) {
        const watermarks = this.watermarks[type];
        if (!watermarks) {
            return 'unknown';
        ***REMOVED***
        if (value < watermarks[0]) {
            return 'low';
        ***REMOVED***
        if (value >= watermarks[1]) {
            return 'high';
        ***REMOVED***
        return 'medium';
    ***REMOVED***

    /**
     * returns an XML writer for the supplied content writer
     * @param {ContentWriter***REMOVED*** contentWriter the content writer to which the returned XML writer
     *  writes data
     * @returns {XMLWriter***REMOVED***
     */
    getXMLWriter(contentWriter) {
        return new XMLWriter(contentWriter);
    ***REMOVED***

    /**
     * returns a full visitor given a partial one.
     * @param {Object***REMOVED*** partialVisitor a partial visitor only having the functions of
     *  interest to the caller. These functions are called with a scope that is the
     *  supplied object.
     * @returns {Visitor***REMOVED***
     */
    getVisitor(partialVisitor) {
        return new tree.Visitor(partialVisitor);
    ***REMOVED***

    getTree(name = 'defaultSummarizer') {
        return this._summarizerFactory[name];
    ***REMOVED***
***REMOVED***

Object.defineProperty(Context.prototype, 'writer', {
    enumerable: true,
    get() {
        if (!this.data.writer) {
            this.data.writer = new FileWriter(this.dir);
        ***REMOVED***
        return this.data.writer;
    ***REMOVED***
***REMOVED***);

module.exports = Context;
