// istanbul-lib-report/lib/xml-writer.js
'use strict';
/*
 Copyright 2012-2015, Yahoo Inc.
 Copyrights licensed under the New BSD License. See the accompanying LICENSE file for terms.
 */
const INDENT = '  ';

function attrString(attrs) {
    return Object.entries(attrs || {***REMOVED***)
        .map(([k, v]) => ` ${k***REMOVED***="${v***REMOVED***"`)
        .join('');
***REMOVED***

/**
 * a utility class to produce well-formed, indented XML
 * @param {ContentWriter***REMOVED*** contentWriter the content writer that this utility wraps
 * @constructor
 */
class XMLWriter {
    constructor(contentWriter) {
        this.cw = contentWriter;
        this.stack = [];
    ***REMOVED***

    indent(str) {
        return this.stack.map(() => INDENT).join('') + str;
    ***REMOVED***

    /**
     * writes the opening XML tag with the supplied attributes
     * @param {String***REMOVED*** name tag name
     * @param {Object***REMOVED*** [attrs=null] attrs attributes for the tag
     */
    openTag(name, attrs) {
        const str = this.indent(`<${name + attrString(attrs)***REMOVED***>`);
        this.cw.println(str);
        this.stack.push(name);
    ***REMOVED***

    /**
     * closes an open XML tag.
     * @param {String***REMOVED*** name - tag name to close. This must match the writer's
     *  notion of the tag that is currently open.
     */
    closeTag(name) {
        if (this.stack.length === 0) {
            throw new Error(`Attempt to close tag ${name***REMOVED*** when not opened`);
        ***REMOVED***
        const stashed = this.stack.pop();
        const str = `</${name***REMOVED***>`;

        if (stashed !== name) {
            throw new Error(
                `Attempt to close tag ${name***REMOVED*** when ${stashed***REMOVED*** was the one open`
            );
        ***REMOVED***
        this.cw.println(this.indent(str));
    ***REMOVED***

    /**
     * writes a tag and its value opening and closing it at the same time
     * @param {String***REMOVED*** name tag name
     * @param {Object***REMOVED*** [attrs=null] attrs tag attributes
     * @param {String***REMOVED*** [content=null] content optional tag content
     */
    inlineTag(name, attrs, content) {
        let str = '<' + name + attrString(attrs);
        if (content) {
            str += `>${content***REMOVED***</${name***REMOVED***>`;
        ***REMOVED*** else {
            str += '/>';
        ***REMOVED***
        str = this.indent(str);
        this.cw.println(str);
    ***REMOVED***

    /**
     * closes all open tags and ends the document
     */
    closeAll() {
        this.stack
            .slice()
            .reverse()
            .forEach(name => {
                this.closeTag(name);
            ***REMOVED***);
    ***REMOVED***
***REMOVED***

module.exports = XMLWriter;
