// @jridgewell/source-map/dist/source-map.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true ***REMOVED***);

var traceMapping = require('@jridgewell/trace-mapping');
var genMapping = require('@jridgewell/gen-mapping');

class SourceMapConsumer {
    constructor(map, mapUrl) {
        const trace = (this._map = new traceMapping.AnyMap(map, mapUrl));
        this.file = trace.file;
        this.names = trace.names;
        this.sourceRoot = trace.sourceRoot;
        this.sources = trace.resolvedSources;
        this.sourcesContent = trace.sourcesContent;
        this.version = trace.version;
    ***REMOVED***
    static fromSourceMap(map, mapUrl) {
        // This is more performant if we receive
        // a @jridgewell/source-map SourceMapGenerator
        if (map.toDecodedMap) {
            return new SourceMapConsumer(map.toDecodedMap(), mapUrl);
        ***REMOVED***
        // This is a fallback for `source-map` and `source-map-js`
        return new SourceMapConsumer(map.toJSON(), mapUrl);
    ***REMOVED***
    get mappings() {
        return traceMapping.encodedMappings(this._map);
    ***REMOVED***
    originalPositionFor(needle) {
        return traceMapping.originalPositionFor(this._map, needle);
    ***REMOVED***
    generatedPositionFor(originalPosition) {
        return traceMapping.generatedPositionFor(this._map, originalPosition);
    ***REMOVED***
    allGeneratedPositionsFor(originalPosition) {
        return traceMapping.allGeneratedPositionsFor(this._map, originalPosition);
    ***REMOVED***
    hasContentsOfAllSources() {
        if (!this.sourcesContent || this.sourcesContent.length !== this.sources.length) {
            return false;
        ***REMOVED***
        for (const content of this.sourcesContent) {
            if (content == null) {
                return false;
            ***REMOVED***
        ***REMOVED***
        return true;
    ***REMOVED***
    sourceContentFor(source, nullOnMissing) {
        const sourceContent = traceMapping.sourceContentFor(this._map, source);
        if (sourceContent != null) {
            return sourceContent;
        ***REMOVED***
        if (nullOnMissing) {
            return null;
        ***REMOVED***
        throw new Error(`"${source***REMOVED***" is not in the SourceMap.`);
    ***REMOVED***
    eachMapping(callback, context /*, order?: number*/) {
        // order is ignored as @jridgewell/trace-map doesn't implement it
        traceMapping.eachMapping(this._map, context ? callback.bind(context) : callback);
    ***REMOVED***
    destroy() {
        // noop.
    ***REMOVED***
***REMOVED***
class SourceMapGenerator {
    constructor(opts) {
        // TODO :: should this be duck-typed ?
        this._map = opts instanceof genMapping.GenMapping ? opts : new genMapping.GenMapping(opts);
    ***REMOVED***
    static fromSourceMap(consumer) {
        return new SourceMapGenerator(genMapping.fromMap(consumer));
    ***REMOVED***
    addMapping(mapping) {
        genMapping.maybeAddMapping(this._map, mapping);
    ***REMOVED***
    setSourceContent(source, content) {
        genMapping.setSourceContent(this._map, source, content);
    ***REMOVED***
    toJSON() {
        return genMapping.toEncodedMap(this._map);
    ***REMOVED***
    toString() {
        return JSON.stringify(this.toJSON());
    ***REMOVED***
    toDecodedMap() {
        return genMapping.toDecodedMap(this._map);
    ***REMOVED***
***REMOVED***

exports.SourceMapConsumer = SourceMapConsumer;
exports.SourceMapGenerator = SourceMapGenerator;
