// @jridgewell/source-map/dist/source-map.mjs
import { AnyMap, encodedMappings, originalPositionFor, generatedPositionFor, allGeneratedPositionsFor, sourceContentFor, eachMapping ***REMOVED*** from '@jridgewell/trace-mapping';
import { GenMapping, fromMap, maybeAddMapping, setSourceContent, toEncodedMap, toDecodedMap ***REMOVED*** from '@jridgewell/gen-mapping';

class SourceMapConsumer {
    constructor(map, mapUrl) {
        const trace = (this._map = new AnyMap(map, mapUrl));
        this.file = trace.file;
        this.names = trace.names;
        this.sourceRoot = trace.sourceRoot;
        this.sources = trace.resolvedSources;
        this.sourcesContent = trace.sourcesContent;
        this.version = trace.version;
    ***REMOVED***
    static fromSourceMap(map, mapUrl) {
        // This is more performant if we receive
        // a @jridgewell/source-map SourceMapGenerator
        if (map.toDecodedMap) {
            return new SourceMapConsumer(map.toDecodedMap(), mapUrl);
        ***REMOVED***
        // This is a fallback for `source-map` and `source-map-js`
        return new SourceMapConsumer(map.toJSON(), mapUrl);
    ***REMOVED***
    get mappings() {
        return encodedMappings(this._map);
    ***REMOVED***
    originalPositionFor(needle) {
        return originalPositionFor(this._map, needle);
    ***REMOVED***
    generatedPositionFor(originalPosition) {
        return generatedPositionFor(this._map, originalPosition);
    ***REMOVED***
    allGeneratedPositionsFor(originalPosition) {
        return allGeneratedPositionsFor(this._map, originalPosition);
    ***REMOVED***
    hasContentsOfAllSources() {
        if (!this.sourcesContent || this.sourcesContent.length !== this.sources.length) {
            return false;
        ***REMOVED***
        for (const content of this.sourcesContent) {
            if (content == null) {
                return false;
            ***REMOVED***
        ***REMOVED***
        return true;
    ***REMOVED***
    sourceContentFor(source, nullOnMissing) {
        const sourceContent = sourceContentFor(this._map, source);
        if (sourceContent != null) {
            return sourceContent;
        ***REMOVED***
        if (nullOnMissing) {
            return null;
        ***REMOVED***
        throw new Error(`"${source***REMOVED***" is not in the SourceMap.`);
    ***REMOVED***
    eachMapping(callback, context /*, order?: number*/) {
        // order is ignored as @jridgewell/trace-map doesn't implement it
        eachMapping(this._map, context ? callback.bind(context) : callback);
    ***REMOVED***
    destroy() {
        // noop.
    ***REMOVED***
***REMOVED***
class SourceMapGenerator {
    constructor(opts) {
        // TODO :: should this be duck-typed ?
        this._map = opts instanceof GenMapping ? opts : new GenMapping(opts);
    ***REMOVED***
    static fromSourceMap(consumer) {
        return new SourceMapGenerator(fromMap(consumer));
    ***REMOVED***
    addMapping(mapping) {
        maybeAddMapping(this._map, mapping);
    ***REMOVED***
    setSourceContent(source, content) {
        setSourceContent(this._map, source, content);
    ***REMOVED***
    toJSON() {
        return toEncodedMap(this._map);
    ***REMOVED***
    toString() {
        return JSON.stringify(this.toJSON());
    ***REMOVED***
    toDecodedMap() {
        return toDecodedMap(this._map);
    ***REMOVED***
***REMOVED***

export { SourceMapConsumer, SourceMapGenerator ***REMOVED***;
