// webpack/lib/asset/AssetModulesPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Yuta Hiroto @hiroppy
*/

"use strict";

const {
	ASSET_MODULE_TYPE_RESOURCE,
	ASSET_MODULE_TYPE_INLINE,
	ASSET_MODULE_TYPE,
	ASSET_MODULE_TYPE_SOURCE
***REMOVED*** = require("../ModuleTypeConstants");
const { cleverMerge ***REMOVED*** = require("../util/cleverMerge");
const { compareModulesByIdOrIdentifier ***REMOVED*** = require("../util/comparators");
const createSchemaValidation = require("../util/create-schema-validation");
const memoize = require("../util/memoize");

/** @typedef {import("webpack-sources").Source***REMOVED*** Source */
/** @typedef {import("../../declarations/WebpackOptions").AssetParserOptions***REMOVED*** AssetParserOptions */
/** @typedef {import("../Chunk")***REMOVED*** Chunk */
/** @typedef {import("../Compilation").AssetInfo***REMOVED*** AssetInfo */
/** @typedef {import("../Compiler")***REMOVED*** Compiler */
/** @typedef {import("../Module")***REMOVED*** Module */
/** @typedef {import("../Module").BuildInfo***REMOVED*** BuildInfo */
/** @typedef {import("../Module").CodeGenerationResult***REMOVED*** CodeGenerationResult */
/** @typedef {import("../NormalModule")***REMOVED*** NormalModule */

/**
 * @param {string***REMOVED*** name name of definitions
 * @returns {TODO***REMOVED*** definition
 */
const getSchema = name => {
	const { definitions ***REMOVED*** = require("../../schemas/WebpackOptions.json");
	return {
		definitions,
		oneOf: [{ $ref: `#/definitions/${name***REMOVED***` ***REMOVED***]
	***REMOVED***;
***REMOVED***;

const generatorValidationOptions = {
	name: "Asset Modules Plugin",
	baseDataPath: "generator"
***REMOVED***;
const validateGeneratorOptions = {
	asset: createSchemaValidation(
		require("../../schemas/plugins/asset/AssetGeneratorOptions.check.js"),
		() => getSchema("AssetGeneratorOptions"),
		generatorValidationOptions
	),
	"asset/resource": createSchemaValidation(
		require("../../schemas/plugins/asset/AssetResourceGeneratorOptions.check.js"),
		() => getSchema("AssetResourceGeneratorOptions"),
		generatorValidationOptions
	),
	"asset/inline": createSchemaValidation(
		require("../../schemas/plugins/asset/AssetInlineGeneratorOptions.check.js"),
		() => getSchema("AssetInlineGeneratorOptions"),
		generatorValidationOptions
	)
***REMOVED***;

const validateParserOptions = createSchemaValidation(
	require("../../schemas/plugins/asset/AssetParserOptions.check.js"),
	() => getSchema("AssetParserOptions"),
	{
		name: "Asset Modules Plugin",
		baseDataPath: "parser"
	***REMOVED***
);

const getAssetGenerator = memoize(() => require("./AssetGenerator"));
const getAssetParser = memoize(() => require("./AssetParser"));
const getAssetSourceParser = memoize(() => require("./AssetSourceParser"));
const getAssetSourceGenerator = memoize(() =>
	require("./AssetSourceGenerator")
);

const type = ASSET_MODULE_TYPE;
const plugin = "AssetModulesPlugin";

class AssetModulesPlugin {
	/**
	 * Apply the plugin
	 * @param {Compiler***REMOVED*** compiler the compiler instance
	 * @returns {void***REMOVED***
	 */
	apply(compiler) {
		compiler.hooks.compilation.tap(
			plugin,
			(compilation, { normalModuleFactory ***REMOVED***) => {
				normalModuleFactory.hooks.createParser
					.for(ASSET_MODULE_TYPE)
					.tap(plugin, parserOptions => {
						validateParserOptions(parserOptions);
						parserOptions = cleverMerge(
							/** @type {AssetParserOptions***REMOVED*** */
							(compiler.options.module.parser.asset),
							parserOptions
						);

						let dataUrlCondition = parserOptions.dataUrlCondition;
						if (!dataUrlCondition || typeof dataUrlCondition === "object") {
							dataUrlCondition = {
								maxSize: 8096,
								...dataUrlCondition
							***REMOVED***;
						***REMOVED***

						const AssetParser = getAssetParser();

						return new AssetParser(dataUrlCondition);
					***REMOVED***);
				normalModuleFactory.hooks.createParser
					.for(ASSET_MODULE_TYPE_INLINE)
					.tap(plugin, _parserOptions => {
						const AssetParser = getAssetParser();

						return new AssetParser(true);
					***REMOVED***);
				normalModuleFactory.hooks.createParser
					.for(ASSET_MODULE_TYPE_RESOURCE)
					.tap(plugin, _parserOptions => {
						const AssetParser = getAssetParser();

						return new AssetParser(false);
					***REMOVED***);
				normalModuleFactory.hooks.createParser
					.for(ASSET_MODULE_TYPE_SOURCE)
					.tap(plugin, _parserOptions => {
						const AssetSourceParser = getAssetSourceParser();

						return new AssetSourceParser();
					***REMOVED***);

				for (const type of [
					ASSET_MODULE_TYPE,
					ASSET_MODULE_TYPE_INLINE,
					ASSET_MODULE_TYPE_RESOURCE
				]) {
					normalModuleFactory.hooks.createGenerator
						.for(type)
						.tap(plugin, generatorOptions => {
							validateGeneratorOptions[type](generatorOptions);

							let dataUrl;
							if (type !== ASSET_MODULE_TYPE_RESOURCE) {
								dataUrl = generatorOptions.dataUrl;
								if (!dataUrl || typeof dataUrl === "object") {
									dataUrl = {
										encoding: undefined,
										mimetype: undefined,
										...dataUrl
									***REMOVED***;
								***REMOVED***
							***REMOVED***

							let filename;
							let publicPath;
							let outputPath;
							if (type !== ASSET_MODULE_TYPE_INLINE) {
								filename = generatorOptions.filename;
								publicPath = generatorOptions.publicPath;
								outputPath = generatorOptions.outputPath;
							***REMOVED***

							const AssetGenerator = getAssetGenerator();

							return new AssetGenerator(
								compilation.moduleGraph,
								dataUrl,
								filename,
								publicPath,
								outputPath,
								generatorOptions.emit !== false
							);
						***REMOVED***);
				***REMOVED***
				normalModuleFactory.hooks.createGenerator
					.for(ASSET_MODULE_TYPE_SOURCE)
					.tap(plugin, () => {
						const AssetSourceGenerator = getAssetSourceGenerator();

						return new AssetSourceGenerator(compilation.moduleGraph);
					***REMOVED***);

				compilation.hooks.renderManifest.tap(plugin, (result, options) => {
					const { chunkGraph ***REMOVED*** = compilation;
					const { chunk, codeGenerationResults, runtimeTemplate ***REMOVED*** = options;

					const modules = chunkGraph.getOrderedChunkModulesIterableBySourceType(
						chunk,
						ASSET_MODULE_TYPE,
						compareModulesByIdOrIdentifier(chunkGraph)
					);
					if (modules) {
						for (const module of modules) {
							try {
								const codeGenResult = codeGenerationResults.get(
									module,
									chunk.runtime
								);
								const buildInfo = /** @type {BuildInfo***REMOVED*** */ (module.buildInfo);
								const data =
									/** @type {NonNullable<CodeGenerationResult["data"]>***REMOVED*** */
									(codeGenResult.data);
								const errored = module.getNumberOfErrors() > 0;

								/** @type {string***REMOVED*** */
								let entryFilename;
								/** @type {AssetInfo***REMOVED*** */
								let entryInfo;
								/** @type {string***REMOVED*** */
								let entryHash;

								if (errored) {
									const erroredModule = /** @type {NormalModule***REMOVED*** */ (module);
									const AssetGenerator = getAssetGenerator();
									const [fullContentHash, contentHash] =
										AssetGenerator.getFullContentHash(
											erroredModule,
											runtimeTemplate
										);
									const { filename, assetInfo ***REMOVED*** =
										AssetGenerator.getFilenameWithInfo(
											erroredModule,
											{
												filename:
													erroredModule.generatorOptions &&
													erroredModule.generatorOptions.filename,
												outputPath:
													erroredModule.generatorOptions &&
													erroredModule.generatorOptions.outputPath
											***REMOVED***,
											{
												runtime: chunk.runtime,
												runtimeTemplate,
												chunkGraph
											***REMOVED***,
											contentHash
										);
									entryFilename = filename;
									entryInfo = assetInfo;
									entryHash = fullContentHash;
								***REMOVED*** else {
									entryFilename = buildInfo.filename || data.get("filename");
									entryInfo = buildInfo.assetInfo || data.get("assetInfo");
									entryHash =
										buildInfo.fullContentHash || data.get("fullContentHash");
								***REMOVED***

								result.push({
									render: () =>
										/** @type {Source***REMOVED*** */ (codeGenResult.sources.get(type)),
									filename: entryFilename,
									info: entryInfo,
									auxiliary: true,
									identifier: `assetModule${chunkGraph.getModuleId(module)***REMOVED***`,
									hash: entryHash
								***REMOVED***);
							***REMOVED*** catch (err) {
								/** @type {Error***REMOVED*** */ (err).message +=
									`\nduring rendering of asset ${module.identifier()***REMOVED***`;
								throw err;
							***REMOVED***
						***REMOVED***
					***REMOVED***

					return result;
				***REMOVED***);

				compilation.hooks.prepareModuleExecution.tap(
					"AssetModulesPlugin",
					(options, context) => {
						const { codeGenerationResult ***REMOVED*** = options;
						const source = codeGenerationResult.sources.get(ASSET_MODULE_TYPE);
						if (source === undefined) return;
						const data =
							/** @type {NonNullable<CodeGenerationResult["data"]>***REMOVED*** */
							(codeGenerationResult.data);
						context.assets.set(data.get("filename"), {
							source,
							info: data.get("assetInfo")
						***REMOVED***);
					***REMOVED***
				);
			***REMOVED***
		);
	***REMOVED***
***REMOVED***

module.exports = AssetModulesPlugin;
