// expect/build/asymmetricMatchers.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
***REMOVED***);
exports.stringNotMatching =
  exports.stringNotContaining =
  exports.stringMatching =
  exports.stringContaining =
  exports.objectNotContaining =
  exports.objectContaining =
  exports.notCloseTo =
  exports.closeTo =
  exports.arrayNotContaining =
  exports.arrayContaining =
  exports.anything =
  exports.any =
  exports.AsymmetricMatcher =
    void 0;

var matcherUtils = _interopRequireWildcard(require('jest-matcher-utils'));

var _jasmineUtils = require('./jasmineUtils');

var _jestMatchersObject = require('./jestMatchersObject');

var _utils = require('./utils');

function _getRequireWildcardCache(nodeInterop) {
  if (typeof WeakMap !== 'function') return null;
  var cacheBabelInterop = new WeakMap();
  var cacheNodeInterop = new WeakMap();
  return (_getRequireWildcardCache = function (nodeInterop) {
    return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
  ***REMOVED***)(nodeInterop);
***REMOVED***

function _interopRequireWildcard(obj, nodeInterop) {
  if (!nodeInterop && obj && obj.__esModule) {
    return obj;
  ***REMOVED***
  if (obj === null || (typeof obj !== 'object' && typeof obj !== 'function')) {
    return {default: obj***REMOVED***;
  ***REMOVED***
  var cache = _getRequireWildcardCache(nodeInterop);
  if (cache && cache.has(obj)) {
    return cache.get(obj);
  ***REMOVED***
  var newObj = {***REMOVED***;
  var hasPropertyDescriptor =
    Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var key in obj) {
    if (key !== 'default' && Object.prototype.hasOwnProperty.call(obj, key)) {
      var desc = hasPropertyDescriptor
        ? Object.getOwnPropertyDescriptor(obj, key)
        : null;
      if (desc && (desc.get || desc.set)) {
        Object.defineProperty(newObj, key, desc);
      ***REMOVED*** else {
        newObj[key] = obj[key];
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  newObj.default = obj;
  if (cache) {
    cache.set(obj, newObj);
  ***REMOVED***
  return newObj;
***REMOVED***

var global = (function () {
  if (typeof globalThis !== 'undefined') {
    return globalThis;
  ***REMOVED*** else if (typeof global !== 'undefined') {
    return global;
  ***REMOVED*** else if (typeof self !== 'undefined') {
    return self;
  ***REMOVED*** else if (typeof window !== 'undefined') {
    return window;
  ***REMOVED*** else {
    return Function('return this')();
  ***REMOVED***
***REMOVED***)();

var Symbol = global['jest-symbol-do-not-touch'] || global.Symbol;

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    ***REMOVED***);
  ***REMOVED*** else {
    obj[key] = value;
  ***REMOVED***
  return obj;
***REMOVED***

const utils = Object.freeze({
  ...matcherUtils,
  iterableEquality: _utils.iterableEquality,
  subsetEquality: _utils.subsetEquality
***REMOVED***);

class AsymmetricMatcher {
  constructor(sample, inverse = false) {
    _defineProperty(this, '$$typeof', Symbol.for('jest.asymmetricMatcher'));

    this.sample = sample;
    this.inverse = inverse;
  ***REMOVED***

  getMatcherContext() {
    return {
      ...(0, _jestMatchersObject.getState)(),
      equals: _jasmineUtils.equals,
      isNot: this.inverse,
      utils
    ***REMOVED***;
  ***REMOVED***
***REMOVED***

exports.AsymmetricMatcher = AsymmetricMatcher;

class Any extends AsymmetricMatcher {
  constructor(sample) {
    if (typeof sample === 'undefined') {
      throw new TypeError(
        'any() expects to be passed a constructor function. ' +
          'Please pass one or use anything() to match any object.'
      );
    ***REMOVED***

    super(sample);
  ***REMOVED***

  asymmetricMatch(other) {
    if (this.sample == String) {
      return typeof other == 'string' || other instanceof String;
    ***REMOVED***

    if (this.sample == Number) {
      return typeof other == 'number' || other instanceof Number;
    ***REMOVED***

    if (this.sample == Function) {
      return typeof other == 'function' || other instanceof Function;
    ***REMOVED***

    if (this.sample == Boolean) {
      return typeof other == 'boolean' || other instanceof Boolean;
    ***REMOVED***

    if (this.sample == BigInt) {
      return typeof other == 'bigint' || other instanceof BigInt;
    ***REMOVED***

    if (this.sample == Symbol) {
      return typeof other == 'symbol' || other instanceof Symbol;
    ***REMOVED***

    if (this.sample == Object) {
      return typeof other == 'object';
    ***REMOVED***

    return other instanceof this.sample;
  ***REMOVED***

  toString() {
    return 'Any';
  ***REMOVED***

  getExpectedType() {
    if (this.sample == String) {
      return 'string';
    ***REMOVED***

    if (this.sample == Number) {
      return 'number';
    ***REMOVED***

    if (this.sample == Function) {
      return 'function';
    ***REMOVED***

    if (this.sample == Object) {
      return 'object';
    ***REMOVED***

    if (this.sample == Boolean) {
      return 'boolean';
    ***REMOVED***

    return (0, _jasmineUtils.fnNameFor)(this.sample);
  ***REMOVED***

  toAsymmetricMatcher() {
    return 'Any<' + (0, _jasmineUtils.fnNameFor)(this.sample) + '>';
  ***REMOVED***
***REMOVED***

class Anything extends AsymmetricMatcher {
  asymmetricMatch(other) {
    return !(0, _jasmineUtils.isUndefined)(other) && other !== null;
  ***REMOVED***

  toString() {
    return 'Anything';
  ***REMOVED*** // No getExpectedType method, because it matches either null or undefined.

  toAsymmetricMatcher() {
    return 'Anything';
  ***REMOVED***
***REMOVED***

class ArrayContaining extends AsymmetricMatcher {
  constructor(sample, inverse = false) {
    super(sample, inverse);
  ***REMOVED***

  asymmetricMatch(other) {
    if (!Array.isArray(this.sample)) {
      throw new Error(
        `You must provide an array to ${this.toString()***REMOVED***, not '` +
          typeof this.sample +
          "'."
      );
    ***REMOVED***

    const result =
      this.sample.length === 0 ||
      (Array.isArray(other) &&
        this.sample.every(item =>
          other.some(another => (0, _jasmineUtils.equals)(item, another))
        ));
    return this.inverse ? !result : result;
  ***REMOVED***

  toString() {
    return `Array${this.inverse ? 'Not' : ''***REMOVED***Containing`;
  ***REMOVED***

  getExpectedType() {
    return 'array';
  ***REMOVED***
***REMOVED***

class ObjectContaining extends AsymmetricMatcher {
  constructor(sample, inverse = false) {
    super(sample, inverse);
  ***REMOVED***

  asymmetricMatch(other) {
    if (typeof this.sample !== 'object') {
      throw new Error(
        `You must provide an object to ${this.toString()***REMOVED***, not '` +
          typeof this.sample +
          "'."
      );
    ***REMOVED***

    let result = true;

    for (const property in this.sample) {
      if (
        !(0, _jasmineUtils.hasProperty)(other, property) ||
        !(0, _jasmineUtils.equals)(this.sample[property], other[property])
      ) {
        result = false;
        break;
      ***REMOVED***
    ***REMOVED***

    return this.inverse ? !result : result;
  ***REMOVED***

  toString() {
    return `Object${this.inverse ? 'Not' : ''***REMOVED***Containing`;
  ***REMOVED***

  getExpectedType() {
    return 'object';
  ***REMOVED***
***REMOVED***

class StringContaining extends AsymmetricMatcher {
  constructor(sample, inverse = false) {
    if (!(0, _jasmineUtils.isA)('String', sample)) {
      throw new Error('Expected is not a string');
    ***REMOVED***

    super(sample, inverse);
  ***REMOVED***

  asymmetricMatch(other) {
    const result =
      (0, _jasmineUtils.isA)('String', other) && other.includes(this.sample);
    return this.inverse ? !result : result;
  ***REMOVED***

  toString() {
    return `String${this.inverse ? 'Not' : ''***REMOVED***Containing`;
  ***REMOVED***

  getExpectedType() {
    return 'string';
  ***REMOVED***
***REMOVED***

class StringMatching extends AsymmetricMatcher {
  constructor(sample, inverse = false) {
    if (
      !(0, _jasmineUtils.isA)('String', sample) &&
      !(0, _jasmineUtils.isA)('RegExp', sample)
    ) {
      throw new Error('Expected is not a String or a RegExp');
    ***REMOVED***

    super(new RegExp(sample), inverse);
  ***REMOVED***

  asymmetricMatch(other) {
    const result =
      (0, _jasmineUtils.isA)('String', other) && this.sample.test(other);
    return this.inverse ? !result : result;
  ***REMOVED***

  toString() {
    return `String${this.inverse ? 'Not' : ''***REMOVED***Matching`;
  ***REMOVED***

  getExpectedType() {
    return 'string';
  ***REMOVED***
***REMOVED***

class CloseTo extends AsymmetricMatcher {
  constructor(sample, precision = 2, inverse = false) {
    if (!(0, _jasmineUtils.isA)('Number', sample)) {
      throw new Error('Expected is not a Number');
    ***REMOVED***

    if (!(0, _jasmineUtils.isA)('Number', precision)) {
      throw new Error('Precision is not a Number');
    ***REMOVED***

    super(sample);

    _defineProperty(this, 'precision', void 0);

    this.inverse = inverse;
    this.precision = precision;
  ***REMOVED***

  asymmetricMatch(other) {
    if (!(0, _jasmineUtils.isA)('Number', other)) {
      return false;
    ***REMOVED***

    let result = false;

    if (other === Infinity && this.sample === Infinity) {
      result = true; // Infinity - Infinity is NaN
    ***REMOVED*** else if (other === -Infinity && this.sample === -Infinity) {
      result = true; // -Infinity - -Infinity is NaN
    ***REMOVED*** else {
      result =
        Math.abs(this.sample - other) < Math.pow(10, -this.precision) / 2;
    ***REMOVED***

    return this.inverse ? !result : result;
  ***REMOVED***

  toString() {
    return `Number${this.inverse ? 'Not' : ''***REMOVED***CloseTo`;
  ***REMOVED***

  getExpectedType() {
    return 'number';
  ***REMOVED***
***REMOVED***

const any = expectedObject => new Any(expectedObject);

exports.any = any;

const anything = () => new Anything();

exports.anything = anything;

const arrayContaining = sample => new ArrayContaining(sample);

exports.arrayContaining = arrayContaining;

const arrayNotContaining = sample => new ArrayContaining(sample, true);

exports.arrayNotContaining = arrayNotContaining;

const objectContaining = sample => new ObjectContaining(sample);

exports.objectContaining = objectContaining;

const objectNotContaining = sample => new ObjectContaining(sample, true);

exports.objectNotContaining = objectNotContaining;

const stringContaining = expected => new StringContaining(expected);

exports.stringContaining = stringContaining;

const stringNotContaining = expected => new StringContaining(expected, true);

exports.stringNotContaining = stringNotContaining;

const stringMatching = expected => new StringMatching(expected);

exports.stringMatching = stringMatching;

const stringNotMatching = expected => new StringMatching(expected, true);

exports.stringNotMatching = stringNotMatching;

const closeTo = (expected, precision) => new CloseTo(expected, precision);

exports.closeTo = closeTo;

const notCloseTo = (expected, precision) =>
  new CloseTo(expected, precision, true);

exports.notCloseTo = notCloseTo;
