// workbox-core/build/workbox-core.dev.js
this.workbox = this.workbox || {***REMOVED***;
this.workbox.core = (function (exports) {
    'use strict';

    try {
      self['workbox:core:6.5.4'] && _();
    ***REMOVED*** catch (e) {***REMOVED***

    /*
      Copyright 2019 Google LLC
      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    const logger = (() => {
      // Don't overwrite this value if it's already set.
      // See https://github.com/GoogleChrome/workbox/pull/2284#issuecomment-560470923
      if (!('__WB_DISABLE_DEV_LOGS' in globalThis)) {
        self.__WB_DISABLE_DEV_LOGS = false;
      ***REMOVED***

      let inGroup = false;
      const methodToColorMap = {
        debug: `#7f8c8d`,
        log: `#2ecc71`,
        warn: `#f39c12`,
        error: `#c0392b`,
        groupCollapsed: `#3498db`,
        groupEnd: null // No colored prefix on groupEnd

      ***REMOVED***;

      const print = function (method, args) {
        if (self.__WB_DISABLE_DEV_LOGS) {
          return;
        ***REMOVED***

        if (method === 'groupCollapsed') {
          // Safari doesn't print all console.groupCollapsed() arguments:
          // https://bugs.webkit.org/show_bug.cgi?id=182754
          if (/^((?!chrome|android).)*safari/i.test(navigator.userAgent)) {
            console[method](...args);
            return;
          ***REMOVED***
        ***REMOVED***

        const styles = [`background: ${methodToColorMap[method]***REMOVED***`, `border-radius: 0.5em`, `color: white`, `font-weight: bold`, `padding: 2px 0.5em`]; // When in a group, the workbox prefix is not displayed.

        const logPrefix = inGroup ? [] : ['%cworkbox', styles.join(';')];
        console[method](...logPrefix, ...args);

        if (method === 'groupCollapsed') {
          inGroup = true;
        ***REMOVED***

        if (method === 'groupEnd') {
          inGroup = false;
        ***REMOVED***
      ***REMOVED***; // eslint-disable-next-line @typescript-eslint/ban-types


      const api = {***REMOVED***;
      const loggerMethods = Object.keys(methodToColorMap);

      for (const key of loggerMethods) {
        const method = key;

        api[method] = (...args) => {
          print(method, args);
        ***REMOVED***;
      ***REMOVED***

      return api;
    ***REMOVED***)();

    /*
      Copyright 2018 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    const messages = {
      'invalid-value': ({
        paramName,
        validValueDescription,
        value
      ***REMOVED***) => {
        if (!paramName || !validValueDescription) {
          throw new Error(`Unexpected input to 'invalid-value' error.`);
        ***REMOVED***

        return `The '${paramName***REMOVED***' parameter was given a value with an ` + `unexpected value. ${validValueDescription***REMOVED*** Received a value of ` + `${JSON.stringify(value)***REMOVED***.`;
      ***REMOVED***,
      'not-an-array': ({
        moduleName,
        className,
        funcName,
        paramName
      ***REMOVED***) => {
        if (!moduleName || !className || !funcName || !paramName) {
          throw new Error(`Unexpected input to 'not-an-array' error.`);
        ***REMOVED***

        return `The parameter '${paramName***REMOVED***' passed into ` + `'${moduleName***REMOVED***.${className***REMOVED***.${funcName***REMOVED***()' must be an array.`;
      ***REMOVED***,
      'incorrect-type': ({
        expectedType,
        paramName,
        moduleName,
        className,
        funcName
      ***REMOVED***) => {
        if (!expectedType || !paramName || !moduleName || !funcName) {
          throw new Error(`Unexpected input to 'incorrect-type' error.`);
        ***REMOVED***

        const classNameStr = className ? `${className***REMOVED***.` : '';
        return `The parameter '${paramName***REMOVED***' passed into ` + `'${moduleName***REMOVED***.${classNameStr***REMOVED***` + `${funcName***REMOVED***()' must be of type ${expectedType***REMOVED***.`;
      ***REMOVED***,
      'incorrect-class': ({
        expectedClassName,
        paramName,
        moduleName,
        className,
        funcName,
        isReturnValueProblem
      ***REMOVED***) => {
        if (!expectedClassName || !moduleName || !funcName) {
          throw new Error(`Unexpected input to 'incorrect-class' error.`);
        ***REMOVED***

        const classNameStr = className ? `${className***REMOVED***.` : '';

        if (isReturnValueProblem) {
          return `The return value from ` + `'${moduleName***REMOVED***.${classNameStr***REMOVED***${funcName***REMOVED***()' ` + `must be an instance of class ${expectedClassName***REMOVED***.`;
        ***REMOVED***

        return `The parameter '${paramName***REMOVED***' passed into ` + `'${moduleName***REMOVED***.${classNameStr***REMOVED***${funcName***REMOVED***()' ` + `must be an instance of class ${expectedClassName***REMOVED***.`;
      ***REMOVED***,
      'missing-a-method': ({
        expectedMethod,
        paramName,
        moduleName,
        className,
        funcName
      ***REMOVED***) => {
        if (!expectedMethod || !paramName || !moduleName || !className || !funcName) {
          throw new Error(`Unexpected input to 'missing-a-method' error.`);
        ***REMOVED***

        return `${moduleName***REMOVED***.${className***REMOVED***.${funcName***REMOVED***() expected the ` + `'${paramName***REMOVED***' parameter to expose a '${expectedMethod***REMOVED***' method.`;
      ***REMOVED***,
      'add-to-cache-list-unexpected-type': ({
        entry
      ***REMOVED***) => {
        return `An unexpected entry was passed to ` + `'workbox-precaching.PrecacheController.addToCacheList()' The entry ` + `'${JSON.stringify(entry)***REMOVED***' isn't supported. You must supply an array of ` + `strings with one or more characters, objects with a url property or ` + `Request objects.`;
      ***REMOVED***,
      'add-to-cache-list-conflicting-entries': ({
        firstEntry,
        secondEntry
      ***REMOVED***) => {
        if (!firstEntry || !secondEntry) {
          throw new Error(`Unexpected input to ` + `'add-to-cache-list-duplicate-entries' error.`);
        ***REMOVED***

        return `Two of the entries passed to ` + `'workbox-precaching.PrecacheController.addToCacheList()' had the URL ` + `${firstEntry***REMOVED*** but different revision details. Workbox is ` + `unable to cache and version the asset correctly. Please remove one ` + `of the entries.`;
      ***REMOVED***,
      'plugin-error-request-will-fetch': ({
        thrownErrorMessage
      ***REMOVED***) => {
        if (!thrownErrorMessage) {
          throw new Error(`Unexpected input to ` + `'plugin-error-request-will-fetch', error.`);
        ***REMOVED***

        return `An error was thrown by a plugins 'requestWillFetch()' method. ` + `The thrown error message was: '${thrownErrorMessage***REMOVED***'.`;
      ***REMOVED***,
      'invalid-cache-name': ({
        cacheNameId,
        value
      ***REMOVED***) => {
        if (!cacheNameId) {
          throw new Error(`Expected a 'cacheNameId' for error 'invalid-cache-name'`);
        ***REMOVED***

        return `You must provide a name containing at least one character for ` + `setCacheDetails({${cacheNameId***REMOVED***: '...'***REMOVED***). Received a value of ` + `'${JSON.stringify(value)***REMOVED***'`;
      ***REMOVED***,
      'unregister-route-but-not-found-with-method': ({
        method
      ***REMOVED***) => {
        if (!method) {
          throw new Error(`Unexpected input to ` + `'unregister-route-but-not-found-with-method' error.`);
        ***REMOVED***

        return `The route you're trying to unregister was not  previously ` + `registered for the method type '${method***REMOVED***'.`;
      ***REMOVED***,
      'unregister-route-route-not-registered': () => {
        return `The route you're trying to unregister was not previously ` + `registered.`;
      ***REMOVED***,
      'queue-replay-failed': ({
        name
      ***REMOVED***) => {
        return `Replaying the background sync queue '${name***REMOVED***' failed.`;
      ***REMOVED***,
      'duplicate-queue-name': ({
        name
      ***REMOVED***) => {
        return `The Queue name '${name***REMOVED***' is already being used. ` + `All instances of backgroundSync.Queue must be given unique names.`;
      ***REMOVED***,
      'expired-test-without-max-age': ({
        methodName,
        paramName
      ***REMOVED***) => {
        return `The '${methodName***REMOVED***()' method can only be used when the ` + `'${paramName***REMOVED***' is used in the constructor.`;
      ***REMOVED***,
      'unsupported-route-type': ({
        moduleName,
        className,
        funcName,
        paramName
      ***REMOVED***) => {
        return `The supplied '${paramName***REMOVED***' parameter was an unsupported type. ` + `Please check the docs for ${moduleName***REMOVED***.${className***REMOVED***.${funcName***REMOVED*** for ` + `valid input types.`;
      ***REMOVED***,
      'not-array-of-class': ({
        value,
        expectedClass,
        moduleName,
        className,
        funcName,
        paramName
      ***REMOVED***) => {
        return `The supplied '${paramName***REMOVED***' parameter must be an array of ` + `'${expectedClass***REMOVED***' objects. Received '${JSON.stringify(value)***REMOVED***,'. ` + `Please check the call to ${moduleName***REMOVED***.${className***REMOVED***.${funcName***REMOVED***() ` + `to fix the issue.`;
      ***REMOVED***,
      'max-entries-or-age-required': ({
        moduleName,
        className,
        funcName
      ***REMOVED***) => {
        return `You must define either config.maxEntries or config.maxAgeSeconds` + `in ${moduleName***REMOVED***.${className***REMOVED***.${funcName***REMOVED***`;
      ***REMOVED***,
      'statuses-or-headers-required': ({
        moduleName,
        className,
        funcName
      ***REMOVED***) => {
        return `You must define either config.statuses or config.headers` + `in ${moduleName***REMOVED***.${className***REMOVED***.${funcName***REMOVED***`;
      ***REMOVED***,
      'invalid-string': ({
        moduleName,
        funcName,
        paramName
      ***REMOVED***) => {
        if (!paramName || !moduleName || !funcName) {
          throw new Error(`Unexpected input to 'invalid-string' error.`);
        ***REMOVED***

        return `When using strings, the '${paramName***REMOVED***' parameter must start with ` + `'http' (for cross-origin matches) or '/' (for same-origin matches). ` + `Please see the docs for ${moduleName***REMOVED***.${funcName***REMOVED***() for ` + `more info.`;
      ***REMOVED***,
      'channel-name-required': () => {
        return `You must provide a channelName to construct a ` + `BroadcastCacheUpdate instance.`;
      ***REMOVED***,
      'invalid-responses-are-same-args': () => {
        return `The arguments passed into responsesAreSame() appear to be ` + `invalid. Please ensure valid Responses are used.`;
      ***REMOVED***,
      'expire-custom-caches-only': () => {
        return `You must provide a 'cacheName' property when using the ` + `expiration plugin with a runtime caching strategy.`;
      ***REMOVED***,
      'unit-must-be-bytes': ({
        normalizedRangeHeader
      ***REMOVED***) => {
        if (!normalizedRangeHeader) {
          throw new Error(`Unexpected input to 'unit-must-be-bytes' error.`);
        ***REMOVED***

        return `The 'unit' portion of the Range header must be set to 'bytes'. ` + `The Range header provided was "${normalizedRangeHeader***REMOVED***"`;
      ***REMOVED***,
      'single-range-only': ({
        normalizedRangeHeader
      ***REMOVED***) => {
        if (!normalizedRangeHeader) {
          throw new Error(`Unexpected input to 'single-range-only' error.`);
        ***REMOVED***

        return `Multiple ranges are not supported. Please use a  single start ` + `value, and optional end value. The Range header provided was ` + `"${normalizedRangeHeader***REMOVED***"`;
      ***REMOVED***,
      'invalid-range-values': ({
        normalizedRangeHeader
      ***REMOVED***) => {
        if (!normalizedRangeHeader) {
          throw new Error(`Unexpected input to 'invalid-range-values' error.`);
        ***REMOVED***

        return `The Range header is missing both start and end values. At least ` + `one of those values is needed. The Range header provided was ` + `"${normalizedRangeHeader***REMOVED***"`;
      ***REMOVED***,
      'no-range-header': () => {
        return `No Range header was found in the Request provided.`;
      ***REMOVED***,
      'range-not-satisfiable': ({
        size,
        start,
        end
      ***REMOVED***) => {
        return `The start (${start***REMOVED***) and end (${end***REMOVED***) values in the Range are ` + `not satisfiable by the cached response, which is ${size***REMOVED*** bytes.`;
      ***REMOVED***,
      'attempt-to-cache-non-get-request': ({
        url,
        method
      ***REMOVED***) => {
        return `Unable to cache '${url***REMOVED***' because it is a '${method***REMOVED***' request and ` + `only 'GET' requests can be cached.`;
      ***REMOVED***,
      'cache-put-with-no-response': ({
        url
      ***REMOVED***) => {
        return `There was an attempt to cache '${url***REMOVED***' but the response was not ` + `defined.`;
      ***REMOVED***,
      'no-response': ({
        url,
        error
      ***REMOVED***) => {
        let message = `The strategy could not generate a response for '${url***REMOVED***'.`;

        if (error) {
          message += ` The underlying error is ${error***REMOVED***.`;
        ***REMOVED***

        return message;
      ***REMOVED***,
      'bad-precaching-response': ({
        url,
        status
      ***REMOVED***) => {
        return `The precaching request for '${url***REMOVED***' failed` + (status ? ` with an HTTP status of ${status***REMOVED***.` : `.`);
      ***REMOVED***,
      'non-precached-url': ({
        url
      ***REMOVED***) => {
        return `createHandlerBoundToURL('${url***REMOVED***') was called, but that URL is ` + `not precached. Please pass in a URL that is precached instead.`;
      ***REMOVED***,
      'add-to-cache-list-conflicting-integrities': ({
        url
      ***REMOVED***) => {
        return `Two of the entries passed to ` + `'workbox-precaching.PrecacheController.addToCacheList()' had the URL ` + `${url***REMOVED*** with different integrity values. Please remove one of them.`;
      ***REMOVED***,
      'missing-precache-entry': ({
        cacheName,
        url
      ***REMOVED***) => {
        return `Unable to find a precached response in ${cacheName***REMOVED*** for ${url***REMOVED***.`;
      ***REMOVED***,
      'cross-origin-copy-response': ({
        origin
      ***REMOVED***) => {
        return `workbox-core.copyResponse() can only be used with same-origin ` + `responses. It was passed a response with origin ${origin***REMOVED***.`;
      ***REMOVED***,
      'opaque-streams-source': ({
        type
      ***REMOVED***) => {
        const message = `One of the workbox-streams sources resulted in an ` + `'${type***REMOVED***' response.`;

        if (type === 'opaqueredirect') {
          return `${message***REMOVED*** Please do not use a navigation request that results ` + `in a redirect as a source.`;
        ***REMOVED***

        return `${message***REMOVED*** Please ensure your sources are CORS-enabled.`;
      ***REMOVED***
    ***REMOVED***;

    /*
      Copyright 2018 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */

    const generatorFunction = (code, details = {***REMOVED***) => {
      const message = messages[code];

      if (!message) {
        throw new Error(`Unable to find message for code '${code***REMOVED***'.`);
      ***REMOVED***

      return message(details);
    ***REMOVED***;

    const messageGenerator = generatorFunction;

    /*
      Copyright 2018 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /**
     * Workbox errors should be thrown with this class.
     * This allows use to ensure the type easily in tests,
     * helps developers identify errors from workbox
     * easily and allows use to optimise error
     * messages correctly.
     *
     * @private
     */

    class WorkboxError extends Error {
      /**
       *
       * @param {string***REMOVED*** errorCode The error code that
       * identifies this particular error.
       * @param {Object=***REMOVED*** details Any relevant arguments
       * that will help developers identify issues should
       * be added as a key on the context object.
       */
      constructor(errorCode, details) {
        const message = messageGenerator(errorCode, details);
        super(message);
        this.name = errorCode;
        this.details = details;
      ***REMOVED***

    ***REMOVED***

    /*
      Copyright 2018 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /*
     * This method throws if the supplied value is not an array.
     * The destructed values are required to produce a meaningful error for users.
     * The destructed and restructured object is so it's clear what is
     * needed.
     */

    const isArray = (value, details) => {
      if (!Array.isArray(value)) {
        throw new WorkboxError('not-an-array', details);
      ***REMOVED***
    ***REMOVED***;

    const hasMethod = (object, expectedMethod, details) => {
      const type = typeof object[expectedMethod];

      if (type !== 'function') {
        details['expectedMethod'] = expectedMethod;
        throw new WorkboxError('missing-a-method', details);
      ***REMOVED***
    ***REMOVED***;

    const isType = (object, expectedType, details) => {
      if (typeof object !== expectedType) {
        details['expectedType'] = expectedType;
        throw new WorkboxError('incorrect-type', details);
      ***REMOVED***
    ***REMOVED***;

    const isInstance = (object, // Need the general type to do the check later.
    // eslint-disable-next-line @typescript-eslint/ban-types
    expectedClass, details) => {
      if (!(object instanceof expectedClass)) {
        details['expectedClassName'] = expectedClass.name;
        throw new WorkboxError('incorrect-class', details);
      ***REMOVED***
    ***REMOVED***;

    const isOneOf = (value, validValues, details) => {
      if (!validValues.includes(value)) {
        details['validValueDescription'] = `Valid values are ${JSON.stringify(validValues)***REMOVED***.`;
        throw new WorkboxError('invalid-value', details);
      ***REMOVED***
    ***REMOVED***;

    const isArrayOfClass = (value, // Need general type to do check later.
    expectedClass, // eslint-disable-line
    details) => {
      const error = new WorkboxError('not-array-of-class', details);

      if (!Array.isArray(value)) {
        throw error;
      ***REMOVED***

      for (const item of value) {
        if (!(item instanceof expectedClass)) {
          throw error;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***;

    const finalAssertExports = {
      hasMethod,
      isArray,
      isInstance,
      isOneOf,
      isType,
      isArrayOfClass
    ***REMOVED***;

    /*
      Copyright 2018 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    // Can't change Function type right now.
    // eslint-disable-next-line @typescript-eslint/ban-types

    const quotaErrorCallbacks = new Set();

    /*
      Copyright 2019 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /**
     * Adds a function to the set of quotaErrorCallbacks that will be executed if
     * there's a quota error.
     *
     * @param {Function***REMOVED*** callback
     * @memberof workbox-core
     */
    // Can't change Function type
    // eslint-disable-next-line @typescript-eslint/ban-types

    function registerQuotaErrorCallback(callback) {
      {
        finalAssertExports.isType(callback, 'function', {
          moduleName: 'workbox-core',
          funcName: 'register',
          paramName: 'callback'
        ***REMOVED***);
      ***REMOVED***

      quotaErrorCallbacks.add(callback);

      {
        logger.log('Registered a callback to respond to quota errors.', callback);
      ***REMOVED***
    ***REMOVED***

    /*
      Copyright 2018 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    const _cacheNameDetails = {
      googleAnalytics: 'googleAnalytics',
      precache: 'precache-v2',
      prefix: 'workbox',
      runtime: 'runtime',
      suffix: typeof registration !== 'undefined' ? registration.scope : ''
    ***REMOVED***;

    const _createCacheName = cacheName => {
      return [_cacheNameDetails.prefix, cacheName, _cacheNameDetails.suffix].filter(value => value && value.length > 0).join('-');
    ***REMOVED***;

    const eachCacheNameDetail = fn => {
      for (const key of Object.keys(_cacheNameDetails)) {
        fn(key);
      ***REMOVED***
    ***REMOVED***;

    const cacheNames$1 = {
      updateDetails: details => {
        eachCacheNameDetail(key => {
          if (typeof details[key] === 'string') {
            _cacheNameDetails[key] = details[key];
          ***REMOVED***
        ***REMOVED***);
      ***REMOVED***,
      getGoogleAnalyticsName: userCacheName => {
        return userCacheName || _createCacheName(_cacheNameDetails.googleAnalytics);
      ***REMOVED***,
      getPrecacheName: userCacheName => {
        return userCacheName || _createCacheName(_cacheNameDetails.precache);
      ***REMOVED***,
      getPrefix: () => {
        return _cacheNameDetails.prefix;
      ***REMOVED***,
      getRuntimeName: userCacheName => {
        return userCacheName || _createCacheName(_cacheNameDetails.runtime);
      ***REMOVED***,
      getSuffix: () => {
        return _cacheNameDetails.suffix;
      ***REMOVED***
    ***REMOVED***;

    /*
      Copyright 2020 Google LLC
      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */

    function stripParams(fullURL, ignoreParams) {
      const strippedURL = new URL(fullURL);

      for (const param of ignoreParams) {
        strippedURL.searchParams.delete(param);
      ***REMOVED***

      return strippedURL.href;
    ***REMOVED***
    /**
     * Matches an item in the cache, ignoring specific URL params. This is similar
     * to the `ignoreSearch` option, but it allows you to ignore just specific
     * params (while continuing to match on the others).
     *
     * @private
     * @param {Cache***REMOVED*** cache
     * @param {Request***REMOVED*** request
     * @param {Object***REMOVED*** matchOptions
     * @param {Array<string>***REMOVED*** ignoreParams
     * @return {Promise<Response|undefined>***REMOVED***
     */


    async function cacheMatchIgnoreParams(cache, request, ignoreParams, matchOptions) {
      const strippedRequestURL = stripParams(request.url, ignoreParams); // If the request doesn't include any ignored params, match as normal.

      if (request.url === strippedRequestURL) {
        return cache.match(request, matchOptions);
      ***REMOVED*** // Otherwise, match by comparing keys


      const keysOptions = Object.assign(Object.assign({***REMOVED***, matchOptions), {
        ignoreSearch: true
      ***REMOVED***);
      const cacheKeys = await cache.keys(request, keysOptions);

      for (const cacheKey of cacheKeys) {
        const strippedCacheKeyURL = stripParams(cacheKey.url, ignoreParams);

        if (strippedRequestURL === strippedCacheKeyURL) {
          return cache.match(cacheKey, matchOptions);
        ***REMOVED***
      ***REMOVED***

      return;
    ***REMOVED***

    /*
      Copyright 2019 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    let supportStatus$1;
    /**
     * A utility function that determines whether the current browser supports
     * constructing a [`ReadableStream`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/ReadableStream)
     * object.
     *
     * @return {boolean***REMOVED*** `true`, if the current browser can successfully
     *     construct a `ReadableStream`, `false` otherwise.
     *
     * @private
     */

    function canConstructReadableStream() {
      if (supportStatus$1 === undefined) {
        // See https://github.com/GoogleChrome/workbox/issues/1473
        try {
          new ReadableStream({
            start() {***REMOVED***

          ***REMOVED***);
          supportStatus$1 = true;
        ***REMOVED*** catch (error) {
          supportStatus$1 = false;
        ***REMOVED***
      ***REMOVED***

      return supportStatus$1;
    ***REMOVED***

    /*
      Copyright 2019 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    let supportStatus;
    /**
     * A utility function that determines whether the current browser supports
     * constructing a new `Response` from a `response.body` stream.
     *
     * @return {boolean***REMOVED*** `true`, if the current browser can successfully
     *     construct a `Response` from a `response.body` stream, `false` otherwise.
     *
     * @private
     */

    function canConstructResponseFromBodyStream() {
      if (supportStatus === undefined) {
        const testResponse = new Response('');

        if ('body' in testResponse) {
          try {
            new Response(testResponse.body);
            supportStatus = true;
          ***REMOVED*** catch (error) {
            supportStatus = false;
          ***REMOVED***
        ***REMOVED***

        supportStatus = false;
      ***REMOVED***

      return supportStatus;
    ***REMOVED***

    /*
      Copyright 2019 Google LLC
      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /**
     * A helper function that prevents a promise from being flagged as unused.
     *
     * @private
     **/

    function dontWaitFor(promise) {
      // Effective no-op.
      void promise.then(() => {***REMOVED***);
    ***REMOVED***

    /*
      Copyright 2018 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /**
     * The Deferred class composes Promises in a way that allows for them to be
     * resolved or rejected from outside the constructor. In most cases promises
     * should be used directly, but Deferreds can be necessary when the logic to
     * resolve a promise must be separate.
     *
     * @private
     */

    class Deferred {
      /**
       * Creates a promise and exposes its resolve and reject functions as methods.
       */
      constructor() {
        this.promise = new Promise((resolve, reject) => {
          this.resolve = resolve;
          this.reject = reject;
        ***REMOVED***);
      ***REMOVED***

    ***REMOVED***

    /*
      Copyright 2018 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /**
     * Runs all of the callback functions, one at a time sequentially, in the order
     * in which they were registered.
     *
     * @memberof workbox-core
     * @private
     */

    async function executeQuotaErrorCallbacks() {
      {
        logger.log(`About to run ${quotaErrorCallbacks.size***REMOVED*** ` + `callbacks to clean up caches.`);
      ***REMOVED***

      for (const callback of quotaErrorCallbacks) {
        await callback();

        {
          logger.log(callback, 'is complete.');
        ***REMOVED***
      ***REMOVED***

      {
        logger.log('Finished running callbacks.');
      ***REMOVED***
    ***REMOVED***

    /*
      Copyright 2018 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */

    const getFriendlyURL = url => {
      const urlObj = new URL(String(url), location.href); // See https://github.com/GoogleChrome/workbox/issues/2323
      // We want to include everything, except for the origin if it's same-origin.

      return urlObj.href.replace(new RegExp(`^${location.origin***REMOVED***`), '');
    ***REMOVED***;

    /*
      Copyright 2019 Google LLC
      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /**
     * Returns a promise that resolves and the passed number of milliseconds.
     * This utility is an async/await-friendly version of `setTimeout`.
     *
     * @param {number***REMOVED*** ms
     * @return {Promise***REMOVED***
     * @private
     */

    function timeout(ms) {
      return new Promise(resolve => setTimeout(resolve, ms));
    ***REMOVED***

    /*
      Copyright 2019 Google LLC
      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    const MAX_RETRY_TIME = 2000;
    /**
     * Returns a promise that resolves to a window client matching the passed
     * `resultingClientId`. For browsers that don't support `resultingClientId`
     * or if waiting for the resulting client to apper takes too long, resolve to
     * `undefined`.
     *
     * @param {string***REMOVED*** [resultingClientId]
     * @return {Promise<Client|undefined>***REMOVED***
     * @private
     */

    async function resultingClientExists(resultingClientId) {
      if (!resultingClientId) {
        return;
      ***REMOVED***

      let existingWindows = await self.clients.matchAll({
        type: 'window'
      ***REMOVED***);
      const existingWindowIds = new Set(existingWindows.map(w => w.id));
      let resultingWindow;
      const startTime = performance.now(); // Only wait up to `MAX_RETRY_TIME` to find a matching client.

      while (performance.now() - startTime < MAX_RETRY_TIME) {
        existingWindows = await self.clients.matchAll({
          type: 'window'
        ***REMOVED***);
        resultingWindow = existingWindows.find(w => {
          if (resultingClientId) {
            // If we have a `resultingClientId`, we can match on that.
            return w.id === resultingClientId;
          ***REMOVED*** else {
            // Otherwise match on finding a window not in `existingWindowIds`.
            return !existingWindowIds.has(w.id);
          ***REMOVED***
        ***REMOVED***);

        if (resultingWindow) {
          break;
        ***REMOVED*** // Sleep for 100ms and retry.


        await timeout(100);
      ***REMOVED***

      return resultingWindow;
    ***REMOVED***

    /*
      Copyright 2020 Google LLC
      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /**
     * A utility method that makes it easier to use `event.waitUntil` with
     * async functions and return the result.
     *
     * @param {ExtendableEvent***REMOVED*** event
     * @param {Function***REMOVED*** asyncFn
     * @return {Function***REMOVED***
     * @private
     */

    function waitUntil(event, asyncFn) {
      const returnPromise = asyncFn();
      event.waitUntil(returnPromise);
      return returnPromise;
    ***REMOVED***

    /*
      Copyright 2018 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */

    var _private = /*#__PURE__*/Object.freeze({
        __proto__: null,
        assert: finalAssertExports,
        cacheMatchIgnoreParams: cacheMatchIgnoreParams,
        cacheNames: cacheNames$1,
        canConstructReadableStream: canConstructReadableStream,
        canConstructResponseFromBodyStream: canConstructResponseFromBodyStream,
        dontWaitFor: dontWaitFor,
        Deferred: Deferred,
        executeQuotaErrorCallbacks: executeQuotaErrorCallbacks,
        getFriendlyURL: getFriendlyURL,
        logger: logger,
        resultingClientExists: resultingClientExists,
        timeout: timeout,
        waitUntil: waitUntil,
        WorkboxError: WorkboxError
    ***REMOVED***);

    /*
      Copyright 2019 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /**
     * Get the current cache names and prefix/suffix used by Workbox.
     *
     * `cacheNames.precache` is used for precached assets,
     * `cacheNames.googleAnalytics` is used by `workbox-google-analytics` to
     * store `analytics.js`, and `cacheNames.runtime` is used for everything else.
     *
     * `cacheNames.prefix` can be used to retrieve just the current prefix value.
     * `cacheNames.suffix` can be used to retrieve just the current suffix value.
     *
     * @return {Object***REMOVED*** An object with `precache`, `runtime`, `prefix`, and
     *     `googleAnalytics` properties.
     *
     * @memberof workbox-core
     */

    const cacheNames = {
      get googleAnalytics() {
        return cacheNames$1.getGoogleAnalyticsName();
      ***REMOVED***,

      get precache() {
        return cacheNames$1.getPrecacheName();
      ***REMOVED***,

      get prefix() {
        return cacheNames$1.getPrefix();
      ***REMOVED***,

      get runtime() {
        return cacheNames$1.getRuntimeName();
      ***REMOVED***,

      get suffix() {
        return cacheNames$1.getSuffix();
      ***REMOVED***

    ***REMOVED***;

    /*
      Copyright 2019 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /**
     * Allows developers to copy a response and modify its `headers`, `status`,
     * or `statusText` values (the values settable via a
     * [`ResponseInit`]{@link https://developer.mozilla.org/en-US/docs/Web/API/Response/Response#Syntax***REMOVED***
     * object in the constructor).
     * To modify these values, pass a function as the second argument. That
     * function will be invoked with a single object with the response properties
     * `{headers, status, statusText***REMOVED***`. The return value of this function will
     * be used as the `ResponseInit` for the new `Response`. To change the values
     * either modify the passed parameter(s) and return it, or return a totally
     * new object.
     *
     * This method is intentionally limited to same-origin responses, regardless of
     * whether CORS was used or not.
     *
     * @param {Response***REMOVED*** response
     * @param {Function***REMOVED*** modifier
     * @memberof workbox-core
     */

    async function copyResponse(response, modifier) {
      let origin = null; // If response.url isn't set, assume it's cross-origin and keep origin null.

      if (response.url) {
        const responseURL = new URL(response.url);
        origin = responseURL.origin;
      ***REMOVED***

      if (origin !== self.location.origin) {
        throw new WorkboxError('cross-origin-copy-response', {
          origin
        ***REMOVED***);
      ***REMOVED***

      const clonedResponse = response.clone(); // Create a fresh `ResponseInit` object by cloning the headers.

      const responseInit = {
        headers: new Headers(clonedResponse.headers),
        status: clonedResponse.status,
        statusText: clonedResponse.statusText
      ***REMOVED***; // Apply any user modifications.

      const modifiedResponseInit = modifier ? modifier(responseInit) : responseInit; // Create the new response from the body stream and `ResponseInit`
      // modifications. Note: not all browsers support the Response.body stream,
      // so fall back to reading the entire body into memory as a blob.

      const body = canConstructResponseFromBodyStream() ? clonedResponse.body : await clonedResponse.blob();
      return new Response(body, modifiedResponseInit);
    ***REMOVED***

    /*
      Copyright 2019 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /**
     * Claim any currently available clients once the service worker
     * becomes active. This is normally used in conjunction with `skipWaiting()`.
     *
     * @memberof workbox-core
     */

    function clientsClaim() {
      self.addEventListener('activate', () => self.clients.claim());
    ***REMOVED***

    /*
      Copyright 2019 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /**
     * Modifies the default cache names used by the Workbox packages.
     * Cache names are generated as `<prefix>-<Cache Name>-<suffix>`.
     *
     * @param {Object***REMOVED*** details
     * @param {Object***REMOVED*** [details.prefix] The string to add to the beginning of
     *     the precache and runtime cache names.
     * @param {Object***REMOVED*** [details.suffix] The string to add to the end of
     *     the precache and runtime cache names.
     * @param {Object***REMOVED*** [details.precache] The cache name to use for precache
     *     caching.
     * @param {Object***REMOVED*** [details.runtime] The cache name to use for runtime caching.
     * @param {Object***REMOVED*** [details.googleAnalytics] The cache name to use for
     *     `workbox-google-analytics` caching.
     *
     * @memberof workbox-core
     */

    function setCacheNameDetails(details) {
      {
        Object.keys(details).forEach(key => {
          finalAssertExports.isType(details[key], 'string', {
            moduleName: 'workbox-core',
            funcName: 'setCacheNameDetails',
            paramName: `details.${key***REMOVED***`
          ***REMOVED***);
        ***REMOVED***);

        if ('precache' in details && details['precache'].length === 0) {
          throw new WorkboxError('invalid-cache-name', {
            cacheNameId: 'precache',
            value: details['precache']
          ***REMOVED***);
        ***REMOVED***

        if ('runtime' in details && details['runtime'].length === 0) {
          throw new WorkboxError('invalid-cache-name', {
            cacheNameId: 'runtime',
            value: details['runtime']
          ***REMOVED***);
        ***REMOVED***

        if ('googleAnalytics' in details && details['googleAnalytics'].length === 0) {
          throw new WorkboxError('invalid-cache-name', {
            cacheNameId: 'googleAnalytics',
            value: details['googleAnalytics']
          ***REMOVED***);
        ***REMOVED***
      ***REMOVED***

      cacheNames$1.updateDetails(details);
    ***REMOVED***

    /*
      Copyright 2019 Google LLC

      Use of this source code is governed by an MIT-style
      license that can be found in the LICENSE file or at
      https://opensource.org/licenses/MIT.
    */
    /**
     * This method is deprecated, and will be removed in Workbox v7.
     *
     * Calling self.skipWaiting() is equivalent, and should be used instead.
     *
     * @memberof workbox-core
     */

    function skipWaiting() {
      // Just call self.skipWaiting() directly.
      // See https://github.com/GoogleChrome/workbox/issues/2525
      {
        logger.warn(`skipWaiting() from workbox-core is no longer recommended ` + `and will be removed in Workbox v7. Using self.skipWaiting() instead ` + `is equivalent.`);
      ***REMOVED***

      void self.skipWaiting();
    ***REMOVED***

    exports._private = _private;
    exports.cacheNames = cacheNames;
    exports.clientsClaim = clientsClaim;
    exports.copyResponse = copyResponse;
    exports.registerQuotaErrorCallback = registerQuotaErrorCallback;
    exports.setCacheNameDetails = setCacheNameDetails;
    exports.skipWaiting = skipWaiting;

    return exports;

***REMOVED***({***REMOVED***));
