// language-tags/lib/Subtag.js
/**
 * @author Matthew Caruana Galizia <mattcg@gmail.com>
 * @license MIT: http://mattcg.mit-license.org/
 * @copyright Copyright (c) 2013, Matthew Caruana Galizia
 */

'use strict';

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; ***REMOVED*** : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; ***REMOVED***, _typeof(obj); ***REMOVED***
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); ***REMOVED*** ***REMOVED***
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); ***REMOVED*** ***REMOVED***
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false ***REMOVED***); return Constructor; ***REMOVED***
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true ***REMOVED***); ***REMOVED*** else { obj[key] = value; ***REMOVED*** return obj; ***REMOVED***
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); ***REMOVED***
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); ***REMOVED*** return (hint === "string" ? String : Number)(input); ***REMOVED***
var index = require('language-subtag-registry/data/json/index.json');
var registry = require('language-subtag-registry/data/json/registry.json');
var Subtag = /*#__PURE__*/function () {
  /**
   * @param {string***REMOVED*** subtag
   * @param {string***REMOVED*** type
   */
  function Subtag(subtag, type) {
    _classCallCheck(this, Subtag);
    var types, i, record;

    // Lowercase for consistency (case is only a formatting convention, not a standard requirement).
    subtag = subtag.toLowerCase();
    type = type.toLowerCase();
    function error(code, message) {
      var err;
      err = new Error(message);
      err.code = code;
      err.subtag = subtag;
      throw err;
    ***REMOVED***
    types = index[subtag];
    if (!types) {
      error(Subtag.ERR_NONEXISTENT, 'Non-existent subtag \'' + subtag + '\'.');
    ***REMOVED***
    i = types[type];
    if (!i && 0 !== i) {
      error(Subtag.ERR_NONEXISTENT, 'Non-existent subtag \'' + subtag + '\' of type \'' + type + '\'.');
    ***REMOVED***
    record = registry[i];
    if (!record.Subtag) {
      error(Subtag.ERR_TAG, '\'' + subtag + '\' is a \'' + type + '\' tag.');
    ***REMOVED***
    this.data = {
      subtag: subtag,
      record: record,
      type: type
    ***REMOVED***;
  ***REMOVED***
  _createClass(Subtag, [{
    key: "type",
    value: function type() {
      return this.data.type;
    ***REMOVED***
  ***REMOVED***, {
    key: "descriptions",
    value: function descriptions() {
      // Every record has one or more descriptions (stored as an array).
      return this.data.record.Description;
    ***REMOVED***
  ***REMOVED***, {
    key: "preferred",
    value: function preferred() {
      var type,
        preferred = this.data.record['Preferred-Value'];
      if (preferred) {
        type = this.data.type;
        if (type === 'extlang') {
          type = 'language';
        ***REMOVED***
        return new Subtag(preferred, type);
      ***REMOVED***
      return null;
    ***REMOVED***
  ***REMOVED***, {
    key: "script",
    value: function script() {
      var script = this.data.record['Suppress-Script'];
      if (script) {
        return new Subtag(script, 'script');
      ***REMOVED***
      return null;
    ***REMOVED***
  ***REMOVED***, {
    key: "scope",
    value: function scope() {
      return this.data.record.Scope || null;
    ***REMOVED***
  ***REMOVED***, {
    key: "deprecated",
    value: function deprecated() {
      return this.data.record.Deprecated || null;
    ***REMOVED***
  ***REMOVED***, {
    key: "added",
    value: function added() {
      return this.data.record.Added;
    ***REMOVED***
  ***REMOVED***, {
    key: "comments",
    value: function comments() {
      // Comments don't always occur for records, so switch to an empty array if missing.
      return this.data.record.Comments || [];
    ***REMOVED***
  ***REMOVED***, {
    key: "format",
    value: function format() {
      var subtag = this.data.subtag;
      switch (this.data.type) {
        case 'region':
          return subtag.toUpperCase();
        case 'script':
          return subtag[0].toUpperCase() + subtag.slice(1);
      ***REMOVED***
      return subtag;
    ***REMOVED***
  ***REMOVED***, {
    key: "toString",
    value: function toString() {
      return this.format();
    ***REMOVED***
  ***REMOVED***]);
  return Subtag;
***REMOVED***();
_defineProperty(Subtag, "ERR_NONEXISTENT", 1);
_defineProperty(Subtag, "ERR_TAG", 2);
module.exports = Subtag;
