// enquirer/lib/types/boolean.js
'use strict';

const Prompt = require('../prompt');
const { isPrimitive, hasColor ***REMOVED*** = require('../utils');

class BooleanPrompt extends Prompt {
  constructor(options) {
    super(options);
    this.cursorHide();
  ***REMOVED***

  async initialize() {
    let initial = await this.resolve(this.initial, this.state);
    this.input = await this.cast(initial);
    await super.initialize();
  ***REMOVED***

  dispatch(ch) {
    if (!this.isValue(ch)) return this.alert();
    this.input = ch;
    return this.submit();
  ***REMOVED***

  format(value) {
    let { styles, state ***REMOVED*** = this;
    return !state.submitted ? styles.primary(value) : styles.success(value);
  ***REMOVED***

  cast(input) {
    return this.isTrue(input);
  ***REMOVED***

  isTrue(input) {
    return /^[ty1]/i.test(input);
  ***REMOVED***

  isFalse(input) {
    return /^[fn0]/i.test(input);
  ***REMOVED***

  isValue(value) {
    return isPrimitive(value) && (this.isTrue(value) || this.isFalse(value));
  ***REMOVED***

  async hint() {
    if (this.state.status === 'pending') {
      let hint = await this.element('hint');
      if (!hasColor(hint)) {
        return this.styles.muted(hint);
      ***REMOVED***
      return hint;
    ***REMOVED***
  ***REMOVED***

  async render() {
    let { input, size ***REMOVED*** = this.state;

    let prefix = await this.prefix();
    let sep = await this.separator();
    let msg = await this.message();
    let hint = this.styles.muted(this.default);

    let promptLine = [prefix, msg, hint, sep].filter(Boolean).join(' ');
    this.state.prompt = promptLine;

    let header = await this.header();
    let value = this.value = this.cast(input);
    let output = await this.format(value);
    let help = (await this.error()) || (await this.hint());
    let footer = await this.footer();

    if (help && !promptLine.includes(help)) output += ' ' + help;
    promptLine += ' ' + output;

    this.clear(size);
    this.write([header, promptLine, footer].filter(Boolean).join('\n'));
    this.restore();
  ***REMOVED***

  set value(value) {
    super.value = value;
  ***REMOVED***
  get value() {
    return this.cast(super.value);
  ***REMOVED***
***REMOVED***

module.exports = BooleanPrompt;
