// enquirer/lib/types/number.js
'use strict';

const StringPrompt = require('./string');

class NumberPrompt extends StringPrompt {
  constructor(options = {***REMOVED***) {
    super({ style: 'number', ...options ***REMOVED***);
    this.min = this.isValue(options.min) ? this.toNumber(options.min) : -Infinity;
    this.max = this.isValue(options.max) ? this.toNumber(options.max) : Infinity;
    this.delay = options.delay != null ? options.delay : 1000;
    this.float = options.float !== false;
    this.round = options.round === true || options.float === false;
    this.major = options.major || 10;
    this.minor = options.minor || 1;
    this.initial = options.initial != null ? options.initial : '';
    this.input = String(this.initial);
    this.cursor = this.input.length;
    this.cursorShow();
  ***REMOVED***

  append(ch) {
    if (!/[-+.]/.test(ch) || (ch === '.' && this.input.includes('.'))) {
      return this.alert('invalid number');
    ***REMOVED***
    return super.append(ch);
  ***REMOVED***

  number(ch) {
    return super.append(ch);
  ***REMOVED***

  next() {
    if (this.input && this.input !== this.initial) return this.alert();
    if (!this.isValue(this.initial)) return this.alert();
    this.input = this.initial;
    this.cursor = String(this.initial).length;
    return this.render();
  ***REMOVED***

  up(number) {
    let step = number || this.minor;
    let num = this.toNumber(this.input);
    if (num > this.max + step) return this.alert();
    this.input = `${num + step***REMOVED***`;
    return this.render();
  ***REMOVED***

  down(number) {
    let step = number || this.minor;
    let num = this.toNumber(this.input);
    if (num < this.min - step) return this.alert();
    this.input = `${num - step***REMOVED***`;
    return this.render();
  ***REMOVED***

  shiftDown() {
    return this.down(this.major);
  ***REMOVED***

  shiftUp() {
    return this.up(this.major);
  ***REMOVED***

  format(input = this.input) {
    if (typeof this.options.format === 'function') {
      return this.options.format.call(this, input);
    ***REMOVED***
    return this.styles.info(input);
  ***REMOVED***

  toNumber(value = '') {
    return this.float ? +value : Math.round(+value);
  ***REMOVED***

  isValue(value) {
    return /^[-+]?[0-9]+((\.)|(\.[0-9]+))?$/.test(value);
  ***REMOVED***

  submit() {
    let value = [this.input, this.initial].find(v => this.isValue(v));
    this.value = this.toNumber(value || 0);
    return super.submit();
  ***REMOVED***
***REMOVED***

module.exports = NumberPrompt;
