// enquirer/lib/types/string.js
'use strict';

const Prompt = require('../prompt');
const keypress = require('../keypress');
const placeholder = require('../placeholder');
const { isPrimitive ***REMOVED*** = require('../utils');

class StringPrompt extends Prompt {
  constructor(options) {
    super(options);
    this.initial = isPrimitive(this.initial) ? String(this.initial) : '';
    if (this.initial) this.cursorHide();
    this.state.prevCursor = 0;
    this.state.clipboard = [];
    this.keypressTimeout =
      this.options.keypressTimeout !== undefined
        ? this.options.keypressTimeout
        : null;
  ***REMOVED***

  async keypress(input, key = input ? keypress(input, {***REMOVED***) : {***REMOVED***) {
    const now = Date.now();
    const elapsed = now - this.lastKeypress;
    this.lastKeypress = now;

    const isEnterKey = key.name === 'return' || key.name === 'enter';
    let prev = this.state.prevKeypress;
    let append;
    this.state.prevKeypress = key;

    if (this.keypressTimeout != null && isEnterKey) {
      if (elapsed < this.keypressTimeout) {
        return this.submit();
      ***REMOVED***

      this.state.multilineBuffer = this.state.multilineBuffer || '';
      this.state.multilineBuffer += input;
      append = true;
      prev = null;
    ***REMOVED***

    if (append || (this.options.multiline && isEnterKey)) {
      if (!prev || prev.name !== 'return') {
        return this.append('\n', key);
      ***REMOVED***
    ***REMOVED***

    return super.keypress(input, key);
  ***REMOVED***

  moveCursor(n) {
    this.cursor += n;
  ***REMOVED***

  reset() {
    this.input = this.value = '';
    this.cursor = 0;
    return this.render();
  ***REMOVED***

  dispatch(ch, key) {
    if (!ch || key.ctrl || key.code) return this.alert();
    this.append(ch);
  ***REMOVED***

  append(ch) {
    let { cursor, input ***REMOVED*** = this.state;
    this.input = `${input***REMOVED***`.slice(0, cursor) + ch + `${input***REMOVED***`.slice(cursor);
    this.moveCursor(String(ch).length);
    this.render();
  ***REMOVED***

  insert(str) {
    this.append(str);
  ***REMOVED***

  delete() {
    let { cursor, input ***REMOVED*** = this.state;
    if (cursor <= 0) return this.alert();
    this.input = `${input***REMOVED***`.slice(0, cursor - 1) + `${input***REMOVED***`.slice(cursor);
    this.moveCursor(-1);
    this.render();
  ***REMOVED***

  deleteForward() {
    let { cursor, input ***REMOVED*** = this.state;
    if (input[cursor] === void 0) return this.alert();
    this.input = `${input***REMOVED***`.slice(0, cursor) + `${input***REMOVED***`.slice(cursor + 1);
    this.render();
  ***REMOVED***

  cutForward() {
    let pos = this.cursor;
    if (this.input.length <= pos) return this.alert();
    this.state.clipboard.push(this.input.slice(pos));
    this.input = this.input.slice(0, pos);
    this.render();
  ***REMOVED***

  cutLeft() {
    let pos = this.cursor;
    if (pos === 0) return this.alert();
    let before = this.input.slice(0, pos);
    let after = this.input.slice(pos);
    let words = before.split(' ');
    this.state.clipboard.push(words.pop());
    this.input = words.join(' ');
    this.cursor = this.input.length;
    this.input += after;
    this.render();
  ***REMOVED***

  paste() {
    if (!this.state.clipboard.length) return this.alert();
    this.insert(this.state.clipboard.pop());
    this.render();
  ***REMOVED***

  toggleCursor() {
    if (this.state.prevCursor) {
      this.cursor = this.state.prevCursor;
      this.state.prevCursor = 0;
    ***REMOVED*** else {
      this.state.prevCursor = this.cursor;
      this.cursor = 0;
    ***REMOVED***
    this.render();
  ***REMOVED***

  first() {
    this.cursor = 0;
    this.render();
  ***REMOVED***

  last() {
    this.cursor = this.input.length - 1;
    this.render();
  ***REMOVED***

  next() {
    let init = this.initial != null ? String(this.initial) : '';
    if (!init || !init.startsWith(this.input)) return this.alert();
    this.input = this.initial;
    this.cursor = this.initial.length;
    this.render();
  ***REMOVED***

  prev() {
    if (!this.input) return this.alert();
    this.reset();
  ***REMOVED***

  backward() {
    return this.left();
  ***REMOVED***

  forward() {
    return this.right();
  ***REMOVED***

  right() {
    if (this.cursor >= this.input.length) return this.alert();
    this.moveCursor(1);
    return this.render();
  ***REMOVED***

  left() {
    if (this.cursor <= 0) return this.alert();
    this.moveCursor(-1);
    return this.render();
  ***REMOVED***

  isValue(value) {
    return !!value;
  ***REMOVED***

  async format(input = this.value) {
    let initial = await this.resolve(this.initial, this.state);
    if (!this.state.submitted) {
      return placeholder(this, { input, initial, pos: this.cursor ***REMOVED***);
    ***REMOVED***
    return this.styles.submitted(input || initial);
  ***REMOVED***

  async render() {
    let size = this.state.size;

    let prefix = await this.prefix();
    let separator = await this.separator();
    let message = await this.message();

    let prompt = [prefix, message, separator].filter(Boolean).join(' ');
    this.state.prompt = prompt;

    let header = await this.header();
    let output = await this.format();
    let help = (await this.error()) || (await this.hint());
    let footer = await this.footer();

    if (help && !output.includes(help)) output += ' ' + help;
    prompt += ' ' + output;

    this.clear(size);
    this.write([header, prompt, footer].filter(Boolean).join('\n'));
    this.restore();
  ***REMOVED***
***REMOVED***

module.exports = StringPrompt;
