// sucrase/dist/util/getImportExportSpecifierInfo.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true***REMOVED***);var _types = require('../parser/tokenizer/types');
















/**
 * Determine information about this named import or named export specifier.
 *
 * This syntax is the `a` from statements like these:
 * import {A***REMOVED*** from "./foo";
 * export {A***REMOVED***;
 * export {A***REMOVED*** from "./foo";
 *
 * As it turns out, we can exactly characterize the syntax meaning by simply
 * counting the number of tokens, which can be from 1 to 4:
 * {A***REMOVED***
 * {type A***REMOVED***
 * {A as B***REMOVED***
 * {type A as B***REMOVED***
 *
 * In the type case, we never actually need the names in practice, so don't get
 * them.
 *
 * TODO: There's some redundancy with the type detection here and the isType
 * flag that's already present on tokens in TS mode. This function could
 * potentially be simplified and/or pushed to the call sites to avoid the object
 * allocation.
 */
 function getImportExportSpecifierInfo(
  tokens,
  index = tokens.currentIndex(),
) {
  let endIndex = index + 1;
  if (isSpecifierEnd(tokens, endIndex)) {
    // import {A***REMOVED***
    const name = tokens.identifierNameAtIndex(index);
    return {
      isType: false,
      leftName: name,
      rightName: name,
      endIndex,
    ***REMOVED***;
  ***REMOVED***
  endIndex++;
  if (isSpecifierEnd(tokens, endIndex)) {
    // import {type A***REMOVED***
    return {
      isType: true,
      leftName: null,
      rightName: null,
      endIndex,
    ***REMOVED***;
  ***REMOVED***
  endIndex++;
  if (isSpecifierEnd(tokens, endIndex)) {
    // import {A as B***REMOVED***
    return {
      isType: false,
      leftName: tokens.identifierNameAtIndex(index),
      rightName: tokens.identifierNameAtIndex(index + 2),
      endIndex,
    ***REMOVED***;
  ***REMOVED***
  endIndex++;
  if (isSpecifierEnd(tokens, endIndex)) {
    // import {type A as B***REMOVED***
    return {
      isType: true,
      leftName: null,
      rightName: null,
      endIndex,
    ***REMOVED***;
  ***REMOVED***
  throw new Error(`Unexpected import/export specifier at ${index***REMOVED***`);
***REMOVED*** exports.default = getImportExportSpecifierInfo;

function isSpecifierEnd(tokens, index) {
  const token = tokens.tokens[index];
  return token.type === _types.TokenType.braceR || token.type === _types.TokenType.comma;
***REMOVED***
