// sucrase/dist/esm/parser/plugins/jsx/index.js
import {
  eat,
  finishToken,
  getTokenFromCode,
  IdentifierRole,
  JSXRole,
  match,
  next,
  skipSpace,
  Token,
***REMOVED*** from "../../tokenizer/index";
import {TokenType as tt***REMOVED*** from "../../tokenizer/types";
import {input, isTypeScriptEnabled, state***REMOVED*** from "../../traverser/base";
import {parseExpression, parseMaybeAssign***REMOVED*** from "../../traverser/expression";
import {expect, unexpected***REMOVED*** from "../../traverser/util";
import {charCodes***REMOVED*** from "../../util/charcodes";
import {IS_IDENTIFIER_CHAR, IS_IDENTIFIER_START***REMOVED*** from "../../util/identifier";
import {tsTryParseJSXTypeArgument***REMOVED*** from "../typescript";

/**
 * Read token with JSX contents.
 *
 * In addition to detecting jsxTagStart and also regular tokens that might be
 * part of an expression, this code detects the start and end of text ranges
 * within JSX children. In order to properly count the number of children, we
 * distinguish jsxText from jsxEmptyText, which is a text range that simplifies
 * to the empty string after JSX whitespace trimming.
 *
 * It turns out that a JSX text range will simplify to the empty string if and
 * only if both of these conditions hold:
 * - The range consists entirely of whitespace characters (only counting space,
 *   tab, \r, and \n).
 * - The range has at least one newline.
 * This can be proven by analyzing any implementation of whitespace trimming,
 * e.g. formatJSXTextLiteral in Sucrase or cleanJSXElementLiteralChild in Babel.
 */
function jsxReadToken() {
  let sawNewline = false;
  let sawNonWhitespace = false;
  while (true) {
    if (state.pos >= input.length) {
      unexpected("Unterminated JSX contents");
      return;
    ***REMOVED***

    const ch = input.charCodeAt(state.pos);
    if (ch === charCodes.lessThan || ch === charCodes.leftCurlyBrace) {
      if (state.pos === state.start) {
        if (ch === charCodes.lessThan) {
          state.pos++;
          finishToken(tt.jsxTagStart);
          return;
        ***REMOVED***
        getTokenFromCode(ch);
        return;
      ***REMOVED***
      if (sawNewline && !sawNonWhitespace) {
        finishToken(tt.jsxEmptyText);
      ***REMOVED*** else {
        finishToken(tt.jsxText);
      ***REMOVED***
      return;
    ***REMOVED***

    // This is part of JSX text.
    if (ch === charCodes.lineFeed) {
      sawNewline = true;
    ***REMOVED*** else if (ch !== charCodes.space && ch !== charCodes.carriageReturn && ch !== charCodes.tab) {
      sawNonWhitespace = true;
    ***REMOVED***
    state.pos++;
  ***REMOVED***
***REMOVED***

function jsxReadString(quote) {
  state.pos++;
  for (;;) {
    if (state.pos >= input.length) {
      unexpected("Unterminated string constant");
      return;
    ***REMOVED***

    const ch = input.charCodeAt(state.pos);
    if (ch === quote) {
      state.pos++;
      break;
    ***REMOVED***
    state.pos++;
  ***REMOVED***
  finishToken(tt.string);
***REMOVED***

// Read a JSX identifier (valid tag or attribute name).
//
// Optimized version since JSX identifiers can't contain
// escape characters and so can be read as single slice.
// Also assumes that first character was already checked
// by isIdentifierStart in readToken.

function jsxReadWord() {
  let ch;
  do {
    if (state.pos > input.length) {
      unexpected("Unexpectedly reached the end of input.");
      return;
    ***REMOVED***
    ch = input.charCodeAt(++state.pos);
  ***REMOVED*** while (IS_IDENTIFIER_CHAR[ch] || ch === charCodes.dash);
  finishToken(tt.jsxName);
***REMOVED***

// Parse next token as JSX identifier
function jsxParseIdentifier() {
  nextJSXTagToken();
***REMOVED***

// Parse namespaced identifier.
function jsxParseNamespacedName(identifierRole) {
  jsxParseIdentifier();
  if (!eat(tt.colon)) {
    // Plain identifier, so this is an access.
    state.tokens[state.tokens.length - 1].identifierRole = identifierRole;
    return;
  ***REMOVED***
  // Process the second half of the namespaced name.
  jsxParseIdentifier();
***REMOVED***

// Parses element name in any form - namespaced, member
// or single identifier.
function jsxParseElementName() {
  const firstTokenIndex = state.tokens.length;
  jsxParseNamespacedName(IdentifierRole.Access);
  let hadDot = false;
  while (match(tt.dot)) {
    hadDot = true;
    nextJSXTagToken();
    jsxParseIdentifier();
  ***REMOVED***
  // For tags like <div> with a lowercase letter and no dots, the name is
  // actually *not* an identifier access, since it's referring to a built-in
  // tag name. Remove the identifier role in this case so that it's not
  // accidentally transformed by the imports transform when preserving JSX.
  if (!hadDot) {
    const firstToken = state.tokens[firstTokenIndex];
    const firstChar = input.charCodeAt(firstToken.start);
    if (firstChar >= charCodes.lowercaseA && firstChar <= charCodes.lowercaseZ) {
      firstToken.identifierRole = null;
    ***REMOVED***
  ***REMOVED***
***REMOVED***

// Parses any type of JSX attribute value.
function jsxParseAttributeValue() {
  switch (state.type) {
    case tt.braceL:
      next();
      parseExpression();
      nextJSXTagToken();
      return;

    case tt.jsxTagStart:
      jsxParseElement();
      nextJSXTagToken();
      return;

    case tt.string:
      nextJSXTagToken();
      return;

    default:
      unexpected("JSX value should be either an expression or a quoted JSX text");
  ***REMOVED***
***REMOVED***

// Parse JSX spread child, after already processing the {
// Does not parse the closing ***REMOVED***
function jsxParseSpreadChild() {
  expect(tt.ellipsis);
  parseExpression();
***REMOVED***

// Parses JSX opening tag starting after "<".
// Returns true if the tag was self-closing.
// Does not parse the last token.
function jsxParseOpeningElement(initialTokenIndex) {
  if (match(tt.jsxTagEnd)) {
    // This is an open-fragment.
    return false;
  ***REMOVED***
  jsxParseElementName();
  if (isTypeScriptEnabled) {
    tsTryParseJSXTypeArgument();
  ***REMOVED***
  let hasSeenPropSpread = false;
  while (!match(tt.slash) && !match(tt.jsxTagEnd) && !state.error) {
    if (eat(tt.braceL)) {
      hasSeenPropSpread = true;
      expect(tt.ellipsis);
      parseMaybeAssign();
      // ***REMOVED***
      nextJSXTagToken();
      continue;
    ***REMOVED***
    if (
      hasSeenPropSpread &&
      state.end - state.start === 3 &&
      input.charCodeAt(state.start) === charCodes.lowercaseK &&
      input.charCodeAt(state.start + 1) === charCodes.lowercaseE &&
      input.charCodeAt(state.start + 2) === charCodes.lowercaseY
    ) {
      state.tokens[initialTokenIndex].jsxRole = JSXRole.KeyAfterPropSpread;
    ***REMOVED***
    jsxParseNamespacedName(IdentifierRole.ObjectKey);
    if (match(tt.eq)) {
      nextJSXTagToken();
      jsxParseAttributeValue();
    ***REMOVED***
  ***REMOVED***
  const isSelfClosing = match(tt.slash);
  if (isSelfClosing) {
    // /
    nextJSXTagToken();
  ***REMOVED***
  return isSelfClosing;
***REMOVED***

// Parses JSX closing tag starting after "</".
// Does not parse the last token.
function jsxParseClosingElement() {
  if (match(tt.jsxTagEnd)) {
    // Fragment syntax, so we immediately have a tag end.
    return;
  ***REMOVED***
  jsxParseElementName();
***REMOVED***

// Parses entire JSX element, including its opening tag
// (starting after "<"), attributes, contents and closing tag.
// Does not parse the last token.
function jsxParseElementAt() {
  const initialTokenIndex = state.tokens.length - 1;
  state.tokens[initialTokenIndex].jsxRole = JSXRole.NoChildren;
  let numExplicitChildren = 0;
  const isSelfClosing = jsxParseOpeningElement(initialTokenIndex);
  if (!isSelfClosing) {
    nextJSXExprToken();
    while (true) {
      switch (state.type) {
        case tt.jsxTagStart:
          nextJSXTagToken();
          if (match(tt.slash)) {
            nextJSXTagToken();
            jsxParseClosingElement();
            // Key after prop spread takes precedence over number of children,
            // since it means we switch to createElement, which doesn't care
            // about number of children.
            if (state.tokens[initialTokenIndex].jsxRole !== JSXRole.KeyAfterPropSpread) {
              if (numExplicitChildren === 1) {
                state.tokens[initialTokenIndex].jsxRole = JSXRole.OneChild;
              ***REMOVED*** else if (numExplicitChildren > 1) {
                state.tokens[initialTokenIndex].jsxRole = JSXRole.StaticChildren;
              ***REMOVED***
            ***REMOVED***
            return;
          ***REMOVED***
          numExplicitChildren++;
          jsxParseElementAt();
          nextJSXExprToken();
          break;

        case tt.jsxText:
          numExplicitChildren++;
          nextJSXExprToken();
          break;

        case tt.jsxEmptyText:
          nextJSXExprToken();
          break;

        case tt.braceL:
          next();
          if (match(tt.ellipsis)) {
            jsxParseSpreadChild();
            nextJSXExprToken();
            // Spread children are a mechanism to explicitly mark children as
            // static, so count it as 2 children to satisfy the "more than one
            // child" condition.
            numExplicitChildren += 2;
          ***REMOVED*** else {
            // If we see {***REMOVED***, this is an empty pseudo-expression that doesn't
            // count as a child.
            if (!match(tt.braceR)) {
              numExplicitChildren++;
              parseExpression();
            ***REMOVED***
            nextJSXExprToken();
          ***REMOVED***

          break;

        // istanbul ignore next - should never happen
        default:
          unexpected();
          return;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***

// Parses entire JSX element from current position.
// Does not parse the last token.
export function jsxParseElement() {
  nextJSXTagToken();
  jsxParseElementAt();
***REMOVED***

// ==================================
// Overrides
// ==================================

export function nextJSXTagToken() {
  state.tokens.push(new Token());
  skipSpace();
  state.start = state.pos;
  const code = input.charCodeAt(state.pos);

  if (IS_IDENTIFIER_START[code]) {
    jsxReadWord();
  ***REMOVED*** else if (code === charCodes.quotationMark || code === charCodes.apostrophe) {
    jsxReadString(code);
  ***REMOVED*** else {
    // The following tokens are just one character each.
    ++state.pos;
    switch (code) {
      case charCodes.greaterThan:
        finishToken(tt.jsxTagEnd);
        break;
      case charCodes.lessThan:
        finishToken(tt.jsxTagStart);
        break;
      case charCodes.slash:
        finishToken(tt.slash);
        break;
      case charCodes.equalsTo:
        finishToken(tt.eq);
        break;
      case charCodes.leftCurlyBrace:
        finishToken(tt.braceL);
        break;
      case charCodes.dot:
        finishToken(tt.dot);
        break;
      case charCodes.colon:
        finishToken(tt.colon);
        break;
      default:
        unexpected();
    ***REMOVED***
  ***REMOVED***
***REMOVED***

function nextJSXExprToken() {
  state.tokens.push(new Token());
  state.start = state.pos;
  jsxReadToken();
***REMOVED***
