// aria-query/lib/etc/roles/literal/buttonRole.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var buttonRole = {
  abstract: false,
  accessibleNameRequired: true,
  baseConcepts: [],
  childrenPresentational: true,
  nameFrom: ['author', 'contents'],
  prohibitedProps: [],
  props: {
    'aria-disabled': null,
    'aria-expanded': null,
    'aria-haspopup': null,
    'aria-pressed': null
  ***REMOVED***,
  relatedConcepts: [{
    concept: {
      attributes: [{
        name: 'type',
        value: 'button'
      ***REMOVED***],
      name: 'input'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      attributes: [{
        name: 'type',
        value: 'image'
      ***REMOVED***],
      name: 'input'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      attributes: [{
        name: 'type',
        value: 'reset'
      ***REMOVED***],
      name: 'input'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      attributes: [{
        name: 'type',
        value: 'submit'
      ***REMOVED***],
      name: 'input'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      name: 'button'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      name: 'trigger'
    ***REMOVED***,
    module: 'XForms'
  ***REMOVED***],
  requireContextRole: [],
  requiredContextRole: [],
  requiredOwnedElements: [],
  requiredProps: {***REMOVED***,
  superClass: [['roletype', 'widget', 'command']]
***REMOVED***;
var _default = exports.default = buttonRole;
