// aria-query/lib/etc/roles/literal/comboboxRole.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var comboboxRole = {
  abstract: false,
  accessibleNameRequired: true,
  baseConcepts: [],
  childrenPresentational: false,
  nameFrom: ['author'],
  prohibitedProps: [],
  props: {
    'aria-activedescendant': null,
    'aria-autocomplete': null,
    'aria-errormessage': null,
    'aria-invalid': null,
    'aria-readonly': null,
    'aria-required': null,
    'aria-expanded': 'false',
    'aria-haspopup': 'listbox'
  ***REMOVED***,
  relatedConcepts: [{
    concept: {
      attributes: [{
        constraints: ['set'],
        name: 'list'
      ***REMOVED***, {
        name: 'type',
        value: 'email'
      ***REMOVED***],
      name: 'input'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      attributes: [{
        constraints: ['set'],
        name: 'list'
      ***REMOVED***, {
        name: 'type',
        value: 'search'
      ***REMOVED***],
      name: 'input'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      attributes: [{
        constraints: ['set'],
        name: 'list'
      ***REMOVED***, {
        name: 'type',
        value: 'tel'
      ***REMOVED***],
      name: 'input'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      attributes: [{
        constraints: ['set'],
        name: 'list'
      ***REMOVED***, {
        name: 'type',
        value: 'text'
      ***REMOVED***],
      name: 'input'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      attributes: [{
        constraints: ['set'],
        name: 'list'
      ***REMOVED***, {
        name: 'type',
        value: 'url'
      ***REMOVED***],
      name: 'input'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      attributes: [{
        constraints: ['set'],
        name: 'list'
      ***REMOVED***, {
        name: 'type',
        value: 'url'
      ***REMOVED***],
      name: 'input'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      attributes: [{
        constraints: ['undefined'],
        name: 'multiple'
      ***REMOVED***, {
        constraints: ['undefined'],
        name: 'size'
      ***REMOVED***],
      constraints: ['the multiple attribute is not set and the size attribute does not have a value greater than 1'],
      name: 'select'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      name: 'select'
    ***REMOVED***,
    module: 'XForms'
  ***REMOVED***],
  requireContextRole: [],
  requiredContextRole: [],
  requiredOwnedElements: [],
  requiredProps: {
    'aria-controls': null,
    'aria-expanded': 'false'
  ***REMOVED***,
  superClass: [['roletype', 'widget', 'input']]
***REMOVED***;
var _default = exports.default = comboboxRole;
