// aria-query/lib/etc/roles/literal/complementaryRole.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var complementaryRole = {
  abstract: false,
  accessibleNameRequired: false,
  baseConcepts: [],
  childrenPresentational: false,
  nameFrom: ['author'],
  prohibitedProps: [],
  props: {***REMOVED***,
  relatedConcepts: [{
    concept: {
      constraints: ['scoped to the body element', 'scoped to the main element'],
      name: 'aside'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      attributes: [{
        constraints: ['set'],
        name: 'aria-label'
      ***REMOVED***],
      constraints: ['scoped to a sectioning content element', 'scoped to a sectioning root element other than body'],
      name: 'aside'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      attributes: [{
        constraints: ['set'],
        name: 'aria-labelledby'
      ***REMOVED***],
      constraints: ['scoped to a sectioning content element', 'scoped to a sectioning root element other than body'],
      name: 'aside'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***],
  requireContextRole: [],
  requiredContextRole: [],
  requiredOwnedElements: [],
  requiredProps: {***REMOVED***,
  superClass: [['roletype', 'structure', 'section', 'landmark']]
***REMOVED***;
var _default = exports.default = complementaryRole;
