// aria-query/lib/etc/roles/literal/formRole.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var formRole = {
  abstract: false,
  accessibleNameRequired: false,
  baseConcepts: [],
  childrenPresentational: false,
  nameFrom: ['author'],
  prohibitedProps: [],
  props: {***REMOVED***,
  relatedConcepts: [{
    concept: {
      attributes: [{
        constraints: ['set'],
        name: 'aria-label'
      ***REMOVED***],
      name: 'form'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      attributes: [{
        constraints: ['set'],
        name: 'aria-labelledby'
      ***REMOVED***],
      name: 'form'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      attributes: [{
        constraints: ['set'],
        name: 'name'
      ***REMOVED***],
      name: 'form'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***],
  requireContextRole: [],
  requiredContextRole: [],
  requiredOwnedElements: [],
  requiredProps: {***REMOVED***,
  superClass: [['roletype', 'structure', 'section', 'landmark']]
***REMOVED***;
var _default = exports.default = formRole;
