// aria-query/lib/etc/roles/literal/graphicsObjectRole.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var graphicsObjectRole = {
  abstract: false,
  accessibleNameRequired: false,
  baseConcepts: [],
  childrenPresentational: false,
  nameFrom: ['author', 'contents'],
  prohibitedProps: [],
  props: {
    'aria-errormessage': null,
    'aria-expanded': null,
    'aria-haspopup': null,
    'aria-invalid': null
  ***REMOVED***,
  relatedConcepts: [{
    module: 'GRAPHICS',
    concept: {
      name: 'graphics-document'
    ***REMOVED***
  ***REMOVED***, {
    module: 'ARIA',
    concept: {
      name: 'group'
    ***REMOVED***
  ***REMOVED***, {
    module: 'ARIA',
    concept: {
      name: 'img'
    ***REMOVED***
  ***REMOVED***, {
    module: 'GRAPHICS',
    concept: {
      name: 'graphics-symbol'
    ***REMOVED***
  ***REMOVED***],
  requireContextRole: [],
  requiredContextRole: [],
  requiredOwnedElements: [],
  requiredProps: {***REMOVED***,
  superClass: [['roletype', 'structure', 'section', 'group']]
***REMOVED***;
var _default = exports.default = graphicsObjectRole;
