// aria-query/lib/etc/roles/literal/headingRole.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var headingRole = {
  abstract: false,
  accessibleNameRequired: true,
  baseConcepts: [],
  childrenPresentational: false,
  nameFrom: ['author', 'contents'],
  prohibitedProps: [],
  props: {
    'aria-level': '2'
  ***REMOVED***,
  relatedConcepts: [{
    concept: {
      name: 'h1'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      name: 'h2'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      name: 'h3'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      name: 'h4'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      name: 'h5'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      name: 'h6'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***],
  requireContextRole: [],
  requiredContextRole: [],
  requiredOwnedElements: [],
  requiredProps: {
    'aria-level': '2'
  ***REMOVED***,
  superClass: [['roletype', 'structure', 'sectionhead']]
***REMOVED***;
var _default = exports.default = headingRole;
