// aria-query/lib/etc/roles/literal/menuitemRole.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var menuitemRole = {
  abstract: false,
  accessibleNameRequired: true,
  baseConcepts: [],
  childrenPresentational: false,
  nameFrom: ['author', 'contents'],
  prohibitedProps: [],
  props: {
    'aria-disabled': null,
    'aria-expanded': null,
    'aria-haspopup': null,
    'aria-posinset': null,
    'aria-setsize': null
  ***REMOVED***,
  relatedConcepts: [{
    concept: {
      name: 'MENU_ITEM'
    ***REMOVED***,
    module: 'JAPI'
  ***REMOVED***, {
    concept: {
      name: 'listitem'
    ***REMOVED***,
    module: 'ARIA'
  ***REMOVED***, {
    concept: {
      name: 'option'
    ***REMOVED***,
    module: 'ARIA'
  ***REMOVED***],
  requireContextRole: ['group', 'menu', 'menubar'],
  requiredContextRole: ['group', 'menu', 'menubar'],
  requiredOwnedElements: [],
  requiredProps: {***REMOVED***,
  superClass: [['roletype', 'widget', 'command']]
***REMOVED***;
var _default = exports.default = menuitemRole;
