// aria-query/lib/etc/roles/literal/menuRole.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var menuRole = {
  abstract: false,
  accessibleNameRequired: false,
  baseConcepts: [],
  childrenPresentational: false,
  nameFrom: ['author'],
  prohibitedProps: [],
  props: {
    'aria-orientation': 'vertical'
  ***REMOVED***,
  relatedConcepts: [{
    concept: {
      name: 'MENU'
    ***REMOVED***,
    module: 'JAPI'
  ***REMOVED***, {
    concept: {
      name: 'list'
    ***REMOVED***,
    module: 'ARIA'
  ***REMOVED***, {
    concept: {
      name: 'select'
    ***REMOVED***,
    module: 'XForms'
  ***REMOVED***, {
    concept: {
      name: 'sidebar'
    ***REMOVED***,
    module: 'DTB'
  ***REMOVED***],
  requireContextRole: [],
  requiredContextRole: [],
  requiredOwnedElements: [['menuitem', 'group'], ['menuitemradio', 'group'], ['menuitemcheckbox', 'group'], ['menuitem'], ['menuitemcheckbox'], ['menuitemradio']],
  requiredProps: {***REMOVED***,
  superClass: [['roletype', 'widget', 'composite', 'select'], ['roletype', 'structure', 'section', 'group', 'select']]
***REMOVED***;
var _default = exports.default = menuRole;
