// aria-query/lib/etc/roles/literal/regionRole.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var regionRole = {
  abstract: false,
  accessibleNameRequired: true,
  baseConcepts: [],
  childrenPresentational: false,
  nameFrom: ['author'],
  prohibitedProps: [],
  props: {***REMOVED***,
  relatedConcepts: [{
    concept: {
      attributes: [{
        constraints: ['set'],
        name: 'aria-label'
      ***REMOVED***],
      name: 'section'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      attributes: [{
        constraints: ['set'],
        name: 'aria-labelledby'
      ***REMOVED***],
      name: 'section'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      name: 'Device Independence Glossart perceivable unit'
    ***REMOVED***
  ***REMOVED***],
  requireContextRole: [],
  requiredContextRole: [],
  requiredOwnedElements: [],
  requiredProps: {***REMOVED***,
  superClass: [['roletype', 'structure', 'section', 'landmark']]
***REMOVED***;
var _default = exports.default = regionRole;
