// aria-query/lib/etc/roles/literal/rowRole.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var rowRole = {
  abstract: false,
  accessibleNameRequired: false,
  baseConcepts: [],
  childrenPresentational: false,
  nameFrom: ['author', 'contents'],
  prohibitedProps: [],
  props: {
    'aria-colindex': null,
    'aria-expanded': null,
    'aria-level': null,
    'aria-posinset': null,
    'aria-rowindex': null,
    'aria-selected': null,
    'aria-setsize': null
  ***REMOVED***,
  relatedConcepts: [{
    concept: {
      name: 'tr'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***],
  requireContextRole: ['grid', 'rowgroup', 'table', 'treegrid'],
  requiredContextRole: ['grid', 'rowgroup', 'table', 'treegrid'],
  requiredOwnedElements: [['cell'], ['columnheader'], ['gridcell'], ['rowheader']],
  requiredProps: {***REMOVED***,
  superClass: [['roletype', 'structure', 'section', 'group'], ['roletype', 'widget']]
***REMOVED***;
var _default = exports.default = rowRole;
