// aria-query/lib/etc/roles/literal/sliderRole.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var sliderRole = {
  abstract: false,
  accessibleNameRequired: true,
  baseConcepts: [],
  childrenPresentational: true,
  nameFrom: ['author'],
  prohibitedProps: [],
  props: {
    'aria-errormessage': null,
    'aria-haspopup': null,
    'aria-invalid': null,
    'aria-readonly': null,
    'aria-valuetext': null,
    'aria-orientation': 'horizontal',
    'aria-valuemax': '100',
    'aria-valuemin': '0'
  ***REMOVED***,
  relatedConcepts: [{
    concept: {
      attributes: [{
        name: 'type',
        value: 'range'
      ***REMOVED***],
      name: 'input'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***],
  requireContextRole: [],
  requiredContextRole: [],
  requiredOwnedElements: [],
  requiredProps: {
    'aria-valuenow': null
  ***REMOVED***,
  superClass: [['roletype', 'widget', 'input'], ['roletype', 'structure', 'range']]
***REMOVED***;
var _default = exports.default = sliderRole;
