// aria-query/lib/etc/roles/literal/textboxRole.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
var textboxRole = {
  abstract: false,
  accessibleNameRequired: true,
  baseConcepts: [],
  childrenPresentational: false,
  nameFrom: ['author'],
  prohibitedProps: [],
  props: {
    'aria-activedescendant': null,
    'aria-autocomplete': null,
    'aria-errormessage': null,
    'aria-haspopup': null,
    'aria-invalid': null,
    'aria-multiline': null,
    'aria-placeholder': null,
    'aria-readonly': null,
    'aria-required': null
  ***REMOVED***,
  relatedConcepts: [{
    concept: {
      attributes: [{
        constraints: ['undefined'],
        name: 'type'
      ***REMOVED***, {
        constraints: ['undefined'],
        name: 'list'
      ***REMOVED***],
      constraints: ['the list attribute is not set'],
      name: 'input'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      attributes: [{
        constraints: ['undefined'],
        name: 'list'
      ***REMOVED***, {
        name: 'type',
        value: 'email'
      ***REMOVED***],
      constraints: ['the list attribute is not set'],
      name: 'input'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      attributes: [{
        constraints: ['undefined'],
        name: 'list'
      ***REMOVED***, {
        name: 'type',
        value: 'tel'
      ***REMOVED***],
      constraints: ['the list attribute is not set'],
      name: 'input'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      attributes: [{
        constraints: ['undefined'],
        name: 'list'
      ***REMOVED***, {
        name: 'type',
        value: 'text'
      ***REMOVED***],
      constraints: ['the list attribute is not set'],
      name: 'input'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      attributes: [{
        constraints: ['undefined'],
        name: 'list'
      ***REMOVED***, {
        name: 'type',
        value: 'url'
      ***REMOVED***],
      constraints: ['the list attribute is not set'],
      name: 'input'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***, {
    concept: {
      name: 'input'
    ***REMOVED***,
    module: 'XForms'
  ***REMOVED***, {
    concept: {
      name: 'textarea'
    ***REMOVED***,
    module: 'HTML'
  ***REMOVED***],
  requireContextRole: [],
  requiredContextRole: [],
  requiredOwnedElements: [],
  requiredProps: {***REMOVED***,
  superClass: [['roletype', 'widget', 'input']]
***REMOVED***;
var _default = exports.default = textboxRole;
