// jsdom/lib/jsdom/living/attributes/Attr-impl.js
"use strict";

const { setAnExistingAttributeValue ***REMOVED*** = require("../attributes.js");
const NodeImpl = require("../nodes/Node-impl.js").implementation;
const { ATTRIBUTE_NODE ***REMOVED*** = require("../node-type.js");

exports.implementation = class AttrImpl extends NodeImpl {
  constructor(globalObject, args, privateData) {
    super(globalObject, args, privateData);

    this._namespace = privateData.namespace !== undefined ? privateData.namespace : null;
    this._namespacePrefix = privateData.namespacePrefix !== undefined ? privateData.namespacePrefix : null;
    this._localName = privateData.localName;
    this._value = privateData.value !== undefined ? privateData.value : "";
    this._element = privateData.element !== undefined ? privateData.element : null;

    this.nodeType = ATTRIBUTE_NODE;
    this.specified = true;
  ***REMOVED***

  get namespaceURI() {
    return this._namespace;
  ***REMOVED***

  get prefix() {
    return this._namespacePrefix;
  ***REMOVED***

  get localName() {
    return this._localName;
  ***REMOVED***

  get name() {
    return this._qualifiedName;
  ***REMOVED***

  get nodeName() {
    return this._qualifiedName;
  ***REMOVED***

  get value() {
    return this._value;
  ***REMOVED***
  set value(value) {
    setAnExistingAttributeValue(this, value);
  ***REMOVED***

  get ownerElement() {
    return this._element;
  ***REMOVED***

  get _qualifiedName() {
    // https://dom.spec.whatwg.org/#concept-attribute-qualified-name
    if (this._namespacePrefix === null) {
      return this._localName;
    ***REMOVED***

    return this._namespacePrefix + ":" + this._localName;
  ***REMOVED***
***REMOVED***;
