// style-loader/dist/runtime/singletonStyleDomAPI.js
"use strict";

/* istanbul ignore next  */
var replaceText = function replaceText() {
  var textStore = [];
  return function replace(index, replacement) {
    textStore[index] = replacement;
    return textStore.filter(Boolean).join("\n");
  ***REMOVED***;
***REMOVED***();

/* istanbul ignore next  */
function apply(styleElement, index, remove, obj) {
  var css;
  if (remove) {
    css = "";
  ***REMOVED*** else {
    css = "";
    if (obj.supports) {
      css += "@supports (".concat(obj.supports, ") {");
    ***REMOVED***
    if (obj.media) {
      css += "@media ".concat(obj.media, " {");
    ***REMOVED***
    var needLayer = typeof obj.layer !== "undefined";
    if (needLayer) {
      css += "@layer".concat(obj.layer.length > 0 ? " ".concat(obj.layer) : "", " {");
    ***REMOVED***
    css += obj.css;
    if (needLayer) {
      css += "***REMOVED***";
    ***REMOVED***
    if (obj.media) {
      css += "***REMOVED***";
    ***REMOVED***
    if (obj.supports) {
      css += "***REMOVED***";
    ***REMOVED***
  ***REMOVED***

  // For old IE
  /* istanbul ignore if  */
  if (styleElement.styleSheet) {
    styleElement.styleSheet.cssText = replaceText(index, css);
  ***REMOVED*** else {
    var cssNode = document.createTextNode(css);
    var childNodes = styleElement.childNodes;
    if (childNodes[index]) {
      styleElement.removeChild(childNodes[index]);
    ***REMOVED***
    if (childNodes.length) {
      styleElement.insertBefore(cssNode, childNodes[index]);
    ***REMOVED*** else {
      styleElement.appendChild(cssNode);
    ***REMOVED***
  ***REMOVED***
***REMOVED***
var singletonData = {
  singleton: null,
  singletonCounter: 0
***REMOVED***;

/* istanbul ignore next  */
function domAPI(options) {
  if (typeof document === "undefined") return {
    update: function update() {***REMOVED***,
    remove: function remove() {***REMOVED***
  ***REMOVED***;

  // eslint-disable-next-line no-undef,no-use-before-define
  var styleIndex = singletonData.singletonCounter++;
  var styleElement =
  // eslint-disable-next-line no-undef,no-use-before-define
  singletonData.singleton || (
  // eslint-disable-next-line no-undef,no-use-before-define
  singletonData.singleton = options.insertStyleElement(options));
  return {
    update: function update(obj) {
      apply(styleElement, styleIndex, false, obj);
    ***REMOVED***,
    remove: function remove(obj) {
      apply(styleElement, styleIndex, true, obj);
    ***REMOVED***
  ***REMOVED***;
***REMOVED***
module.exports = domAPI;
