// yaml/browser/dist/schema/Schema.js
import { MAP, SCALAR, SEQ ***REMOVED*** from '../nodes/identity.js';
import { map ***REMOVED*** from './common/map.js';
import { seq ***REMOVED*** from './common/seq.js';
import { string ***REMOVED*** from './common/string.js';
import { getTags, coreKnownTags ***REMOVED*** from './tags.js';

const sortMapEntriesByKey = (a, b) => a.key < b.key ? -1 : a.key > b.key ? 1 : 0;
class Schema {
    constructor({ compat, customTags, merge, resolveKnownTags, schema, sortMapEntries, toStringDefaults ***REMOVED***) {
        this.compat = Array.isArray(compat)
            ? getTags(compat, 'compat')
            : compat
                ? getTags(null, compat)
                : null;
        this.name = (typeof schema === 'string' && schema) || 'core';
        this.knownTags = resolveKnownTags ? coreKnownTags : {***REMOVED***;
        this.tags = getTags(customTags, this.name, merge);
        this.toStringOptions = toStringDefaults ?? null;
        Object.defineProperty(this, MAP, { value: map ***REMOVED***);
        Object.defineProperty(this, SCALAR, { value: string ***REMOVED***);
        Object.defineProperty(this, SEQ, { value: seq ***REMOVED***);
        // Used by createMap()
        this.sortMapEntries =
            typeof sortMapEntries === 'function'
                ? sortMapEntries
                : sortMapEntries === true
                    ? sortMapEntriesByKey
                    : null;
    ***REMOVED***
    clone() {
        const copy = Object.create(Schema.prototype, Object.getOwnPropertyDescriptors(this));
        copy.tags = this.tags.slice();
        return copy;
    ***REMOVED***
***REMOVED***

export { Schema ***REMOVED***;
