// @svgr/plugin-jsx/lib/index.js
"use strict";

exports.__esModule = true;
exports.default = jsxPlugin;

var _svgParser = require("svg-parser");

var _hastUtilToBabelAst = _interopRequireDefault(require("@svgr/hast-util-to-babel-ast"));

var _core = require("@babel/core");

var _babelPreset = _interopRequireDefault(require("@svgr/babel-preset"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj ***REMOVED***; ***REMOVED***

function jsxPlugin(code, config, state) {
  const filePath = state.filePath || 'unknown';
  const hastTree = (0, _svgParser.parse)(code);
  const babelTree = (0, _hastUtilToBabelAst.default)(hastTree);
  const {
    code: generatedCode
  ***REMOVED*** = (0, _core.transformFromAstSync)(babelTree, code, {
    caller: {
      name: 'svgr'
    ***REMOVED***,
    presets: [(0, _core.createConfigItem)([_babelPreset.default, { ...config,
      state
    ***REMOVED***], {
      type: 'preset'
    ***REMOVED***)],
    filename: filePath,
    babelrc: false,
    configFile: false,
    code: true,
    ast: false,
    inputSourceMap: false,
    ...(config.jsx && config.jsx.babelConfig)
  ***REMOVED***);
  return generatedCode;
***REMOVED***
