// ajv/lib/vocabularies/applicator/additionalProperties.ts
import type {
  CodeKeywordDefinition,
  AddedKeywordDefinition,
  ErrorObject,
  KeywordErrorDefinition,
  AnySchema,
***REMOVED*** from "../../types"
import {allSchemaProperties, usePattern, isOwnProperty***REMOVED*** from "../code"
import {_, nil, or, not, Code, Name***REMOVED*** from "../../compile/codegen"
import N from "../../compile/names"
import type {SubschemaArgs***REMOVED*** from "../../compile/validate/subschema"
import {alwaysValidSchema, schemaRefOrVal, Type***REMOVED*** from "../../compile/util"

export type AdditionalPropertiesError = ErrorObject<
  "additionalProperties",
  {additionalProperty: string***REMOVED***,
  AnySchema
>

const error: KeywordErrorDefinition = {
  message: "must NOT have additional properties",
  params: ({params***REMOVED***) => _`{additionalProperty: ${params.additionalProperty***REMOVED******REMOVED***`,
***REMOVED***

const def: CodeKeywordDefinition & AddedKeywordDefinition = {
  keyword: "additionalProperties",
  type: ["object"],
  schemaType: ["boolean", "object"],
  allowUndefined: true,
  trackErrors: true,
  error,
  code(cxt) {
    const {gen, schema, parentSchema, data, errsCount, it***REMOVED*** = cxt
    /* istanbul ignore if */
    if (!errsCount) throw new Error("ajv implementation error")
    const {allErrors, opts***REMOVED*** = it
    it.props = true
    if (opts.removeAdditional !== "all" && alwaysValidSchema(it, schema)) return
    const props = allSchemaProperties(parentSchema.properties)
    const patProps = allSchemaProperties(parentSchema.patternProperties)
    checkAdditionalProperties()
    cxt.ok(_`${errsCount***REMOVED*** === ${N.errors***REMOVED***`)

    function checkAdditionalProperties(): void {
      gen.forIn("key", data, (key: Name) => {
        if (!props.length && !patProps.length) additionalPropertyCode(key)
        else gen.if(isAdditional(key), () => additionalPropertyCode(key))
      ***REMOVED***)
    ***REMOVED***

    function isAdditional(key: Name): Code {
      let definedProp: Code
      if (props.length > 8) {
        // TODO maybe an option instead of hard-coded 8?
        const propsSchema = schemaRefOrVal(it, parentSchema.properties, "properties")
        definedProp = isOwnProperty(gen, propsSchema as Code, key)
      ***REMOVED*** else if (props.length) {
        definedProp = or(...props.map((p) => _`${key***REMOVED*** === ${p***REMOVED***`))
      ***REMOVED*** else {
        definedProp = nil
      ***REMOVED***
      if (patProps.length) {
        definedProp = or(definedProp, ...patProps.map((p) => _`${usePattern(cxt, p)***REMOVED***.test(${key***REMOVED***)`))
      ***REMOVED***
      return not(definedProp)
    ***REMOVED***

    function deleteAdditional(key: Name): void {
      gen.code(_`delete ${data***REMOVED***[${key***REMOVED***]`)
    ***REMOVED***

    function additionalPropertyCode(key: Name): void {
      if (opts.removeAdditional === "all" || (opts.removeAdditional && schema === false)) {
        deleteAdditional(key)
        return
      ***REMOVED***

      if (schema === false) {
        cxt.setParams({additionalProperty: key***REMOVED***)
        cxt.error()
        if (!allErrors) gen.break()
        return
      ***REMOVED***

      if (typeof schema == "object" && !alwaysValidSchema(it, schema)) {
        const valid = gen.name("valid")
        if (opts.removeAdditional === "failing") {
          applyAdditionalSchema(key, valid, false)
          gen.if(not(valid), () => {
            cxt.reset()
            deleteAdditional(key)
          ***REMOVED***)
        ***REMOVED*** else {
          applyAdditionalSchema(key, valid)
          if (!allErrors) gen.if(not(valid), () => gen.break())
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    function applyAdditionalSchema(key: Name, valid: Name, errors?: false): void {
      const subschema: SubschemaArgs = {
        keyword: "additionalProperties",
        dataProp: key,
        dataPropType: Type.Str,
      ***REMOVED***
      if (errors === false) {
        Object.assign(subschema, {
          compositeRule: true,
          createErrors: false,
          allErrors: false,
        ***REMOVED***)
      ***REMOVED***
      cxt.subschema(subschema, valid)
    ***REMOVED***
  ***REMOVED***,
***REMOVED***

export default def
