// ajv/lib/vocabularies/applicator/properties.ts
import type {CodeKeywordDefinition***REMOVED*** from "../../types"
import {KeywordCxt***REMOVED*** from "../../compile/validate"
import {propertyInData, allSchemaProperties***REMOVED*** from "../code"
import {alwaysValidSchema, toHash, mergeEvaluated***REMOVED*** from "../../compile/util"
import apDef from "./additionalProperties"

const def: CodeKeywordDefinition = {
  keyword: "properties",
  type: "object",
  schemaType: "object",
  code(cxt: KeywordCxt) {
    const {gen, schema, parentSchema, data, it***REMOVED*** = cxt
    if (it.opts.removeAdditional === "all" && parentSchema.additionalProperties === undefined) {
      apDef.code(new KeywordCxt(it, apDef, "additionalProperties"))
    ***REMOVED***
    const allProps = allSchemaProperties(schema)
    for (const prop of allProps) {
      it.definedProperties.add(prop)
    ***REMOVED***
    if (it.opts.unevaluated && allProps.length && it.props !== true) {
      it.props = mergeEvaluated.props(gen, toHash(allProps), it.props)
    ***REMOVED***
    const properties = allProps.filter((p) => !alwaysValidSchema(it, schema[p]))
    if (properties.length === 0) return
    const valid = gen.name("valid")

    for (const prop of properties) {
      if (hasDefault(prop)) {
        applyPropertySchema(prop)
      ***REMOVED*** else {
        gen.if(propertyInData(gen, data, prop, it.opts.ownProperties))
        applyPropertySchema(prop)
        if (!it.allErrors) gen.else().var(valid, true)
        gen.endIf()
      ***REMOVED***
      cxt.it.definedProperties.add(prop)
      cxt.ok(valid)
    ***REMOVED***

    function hasDefault(prop: string): boolean | undefined {
      return it.opts.useDefaults && !it.compositeRule && schema[prop].default !== undefined
    ***REMOVED***

    function applyPropertySchema(prop: string): void {
      cxt.subschema(
        {
          keyword: "properties",
          schemaProp: prop,
          dataProp: prop,
        ***REMOVED***,
        valid
      )
    ***REMOVED***
  ***REMOVED***,
***REMOVED***

export default def
