// postcss-merge-longhand/src/lib/decl/borders.js
'use strict';
const { list ***REMOVED*** = require('postcss');
const stylehacks = require('stylehacks');
const insertCloned = require('../insertCloned.js');
const parseTrbl = require('../parseTrbl.js');
const hasAllProps = require('../hasAllProps.js');
const getDecls = require('../getDecls.js');
const getRules = require('../getRules.js');
const getValue = require('../getValue.js');
const mergeRules = require('../mergeRules.js');
const minifyTrbl = require('../minifyTrbl.js');
const minifyWsc = require('../minifyWsc.js');
const canMerge = require('../canMerge.js');
const trbl = require('../trbl.js');
const isCustomProp = require('../isCustomProp.js');
const canExplode = require('../canExplode.js');
const getLastNode = require('../getLastNode.js');
const parseWsc = require('../parseWsc.js');
const { isValidWsc ***REMOVED*** = require('../validateWsc.js');

const wsc = ['width', 'style', 'color'];
const defaults = ['medium', 'none', 'currentcolor'];
const colorMightRequireFallback =
  /(hsla|rgba|color|hwb|lab|lch|oklab|oklch)\(/i;

/**
 * @param {...string***REMOVED*** parts
 * @return {string***REMOVED***
 */
function borderProperty(...parts) {
  return `border-${parts.join('-')***REMOVED***`;
***REMOVED***
/**
 * @param {string***REMOVED*** value
 * @return {string***REMOVED***
 */
function mapBorderProperty(value) {
  return borderProperty(value);
***REMOVED***

const directions = trbl.map(mapBorderProperty);
const properties = wsc.map(mapBorderProperty);
/** @type {string[]***REMOVED*** */
const directionalProperties = directions.reduce(
  (prev, curr) => prev.concat(wsc.map((prop) => `${curr***REMOVED***-${prop***REMOVED***`)),
  /** @type {string[]***REMOVED*** */ ([])
);

const precedence = [
  ['border'],
  directions.concat(properties),
  directionalProperties,
];

const allProperties = precedence.reduce((a, b) => a.concat(b));

/**
 * @param {string***REMOVED*** prop
 * @return {number | undefined***REMOVED***
 */
function getLevel(prop) {
  for (let i = 0; i < precedence.length; i++) {
    if (precedence[i].includes(prop.toLowerCase())) {
      return i;
    ***REMOVED***
  ***REMOVED***
***REMOVED***

/** @type {(value: string) => boolean***REMOVED*** */
const isValueCustomProp = (value) =>
  value !== undefined && value.search(/var\s*\(\s*--/i) !== -1;

/**
 * @param {string[]***REMOVED*** values
 * @return {boolean***REMOVED***
 */
function canMergeValues(values) {
  return !values.some(isValueCustomProp);
***REMOVED***

/**
 * @param {import('postcss').Declaration***REMOVED*** decl
 * @return {string***REMOVED***
 */
function getColorValue(decl) {
  if (decl.prop.substr(-5) === 'color') {
    return decl.value;
  ***REMOVED***

  return parseWsc(decl.value)[2] || defaults[2];
***REMOVED***

/**
 * @param {[string, string, string]***REMOVED*** values
 * @param {[string, string, string]***REMOVED*** nextValues
 * @return {string[]***REMOVED***
 */
function diffingProps(values, nextValues) {
  return wsc.reduce((prev, curr, i) => {
    if (values[i] === nextValues[i]) {
      return prev;
    ***REMOVED***

    return [...prev, curr];
  ***REMOVED***, /** @type {string[]***REMOVED*** */ ([]));
***REMOVED***

/**
 * @param {{values: [string, string, string], nextValues: [string, string, string], decl: import('postcss').Declaration, nextDecl: import('postcss').Declaration, index: number***REMOVED******REMOVED*** arg
 * @return {void***REMOVED***
 */
function mergeRedundant({ values, nextValues, decl, nextDecl, index ***REMOVED***) {
  if (!canMerge([decl, nextDecl])) {
    return;
  ***REMOVED***

  if (stylehacks.detect(decl) || stylehacks.detect(nextDecl)) {
    return;
  ***REMOVED***

  const diff = diffingProps(values, nextValues);

  if (diff.length !== 1) {
    return;
  ***REMOVED***

  const prop = /** @type {string***REMOVED*** */ (diff.pop());
  const position = wsc.indexOf(prop);

  const prop1 = `${nextDecl.prop***REMOVED***-${prop***REMOVED***`;
  const prop2 = `border-${prop***REMOVED***`;

  let props = parseTrbl(values[position]);

  props[index] = nextValues[position];

  const borderValue2 = values.filter((e, i) => i !== position).join(' ');
  const propValue2 = minifyTrbl(props);

  const origLength = (minifyWsc(decl.value) + nextDecl.prop + nextDecl.value)
    .length;
  const newLength1 =
    decl.value.length + prop1.length + minifyWsc(nextValues[position]).length;
  const newLength2 = borderValue2.length + prop2.length + propValue2.length;

  if (newLength1 < newLength2 && newLength1 < origLength) {
    nextDecl.prop = prop1;
    nextDecl.value = nextValues[position];
  ***REMOVED***

  if (newLength2 < newLength1 && newLength2 < origLength) {
    decl.value = borderValue2;
    nextDecl.prop = prop2;
    nextDecl.value = propValue2;
  ***REMOVED***
***REMOVED***

/**
 * @param {string | string[]***REMOVED*** mapped
 * @return {boolean***REMOVED***
 */
function isCloseEnough(mapped) {
  return (
    (mapped[0] === mapped[1] && mapped[1] === mapped[2]) ||
    (mapped[1] === mapped[2] && mapped[2] === mapped[3]) ||
    (mapped[2] === mapped[3] && mapped[3] === mapped[0]) ||
    (mapped[3] === mapped[0] && mapped[0] === mapped[1])
  );
***REMOVED***

/**
 * @param {string[]***REMOVED*** mapped
 * @return {string[]***REMOVED***
 */
function getDistinctShorthands(mapped) {
  return [...new Set(mapped)];
***REMOVED***
/**
 * @param {import('postcss').Rule***REMOVED*** rule
 * @return {void***REMOVED***
 */
function explode(rule) {
  rule.walkDecls(/^border/i, (decl) => {
    if (!canExplode(decl, false)) {
      return;
    ***REMOVED***

    if (stylehacks.detect(decl)) {
      return;
    ***REMOVED***

    const prop = decl.prop.toLowerCase();

    // border -> border-trbl
    if (prop === 'border') {
      if (isValidWsc(parseWsc(decl.value))) {
        directions.forEach((direction) => {
          insertCloned(
            /** @type {import('postcss').Rule***REMOVED*** */ (decl.parent),
            decl,
            { prop: direction ***REMOVED***
          );
        ***REMOVED***);

        decl.remove();
      ***REMOVED***
    ***REMOVED***

    // border-trbl -> border-trbl-wsc
    if (directions.some((direction) => prop === direction)) {
      let values = parseWsc(decl.value);

      if (isValidWsc(values)) {
        wsc.forEach((d, i) => {
          insertCloned(
            /** @type {import('postcss').Rule***REMOVED*** */ (decl.parent),
            decl,
            {
              prop: `${prop***REMOVED***-${d***REMOVED***`,
              value: values[i] || defaults[i],
            ***REMOVED***
          );
        ***REMOVED***);

        decl.remove();
      ***REMOVED***
    ***REMOVED***

    // border-wsc -> border-trbl-wsc
    wsc.some((style) => {
      if (prop !== borderProperty(style)) {
        return false;
      ***REMOVED***

      if (isCustomProp(decl)) {
        decl.prop = decl.prop.toLowerCase();
        return false;
      ***REMOVED***
      parseTrbl(decl.value).forEach((value, i) => {
        insertCloned(
          /** @type {import('postcss').Rule***REMOVED*** */ (decl.parent),
          decl,
          {
            prop: borderProperty(trbl[i], style),
            value,
          ***REMOVED***
        );
      ***REMOVED***);

      return decl.remove();
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***

/**
 * @param {import('postcss').Rule***REMOVED*** rule
 * @return {void***REMOVED***
 */
function merge(rule) {
  // border-trbl-wsc -> border-trbl
  trbl.forEach((direction) => {
    const prop = borderProperty(direction);

    mergeRules(
      rule,
      wsc.map((style) => borderProperty(direction, style)),
      (rules, lastNode) => {
        if (canMerge(rules, false) && !rules.some(stylehacks.detect)) {
          insertCloned(
            /** @type {import('postcss').Rule***REMOVED*** */ (lastNode.parent),
            lastNode,
            {
              prop,
              value: rules.map(getValue).join(' '),
            ***REMOVED***
          );
          for (const node of rules) {
            node.remove();
          ***REMOVED***

          return true;
        ***REMOVED***
        return false;
      ***REMOVED***
    );
  ***REMOVED***);

  // border-trbl-wsc -> border-wsc
  wsc.forEach((style) => {
    const prop = borderProperty(style);

    mergeRules(
      rule,
      trbl.map((direction) => borderProperty(direction, style)),
      (rules, lastNode) => {
        if (canMerge(rules) && !rules.some(stylehacks.detect)) {
          insertCloned(
            /** @type {import('postcss').Rule***REMOVED*** */ (lastNode.parent),
            lastNode,
            {
              prop,
              value: minifyTrbl(rules.map(getValue).join(' ')),
            ***REMOVED***
          );

          for (const node of rules) {
            node.remove();
          ***REMOVED***

          return true;
        ***REMOVED***
        return false;
      ***REMOVED***
    );
  ***REMOVED***);

  // border-trbl -> border-wsc
  mergeRules(rule, directions, (rules, lastNode) => {
    if (rules.some(stylehacks.detect)) {
      return false;
    ***REMOVED***

    const values = rules.map(({ value ***REMOVED***) => value);

    if (!canMergeValues(values)) {
      return false;
    ***REMOVED***

    const parsed = values.map((value) => parseWsc(value));

    if (!parsed.every(isValidWsc)) {
      return false;
    ***REMOVED***

    wsc.forEach((d, i) => {
      const value = parsed.map((v) => v[i] || defaults[i]);

      if (canMergeValues(value)) {
        insertCloned(
          /** @type {import('postcss').Rule***REMOVED*** */ (lastNode.parent),
          lastNode,
          {
            prop: borderProperty(d),
            value: minifyTrbl(
              /** @type {[string, string, string, string]***REMOVED*** */ (value)
            ),
          ***REMOVED***
        );
      ***REMOVED*** else {
        insertCloned(
          /** @type {import('postcss').Rule***REMOVED*** */ (lastNode.parent),
          lastNode
        );
      ***REMOVED***
    ***REMOVED***);

    for (const node of rules) {
      node.remove();
    ***REMOVED***

    return true;
  ***REMOVED***);

  // border-wsc -> border
  // border-wsc -> border + border-color
  // border-wsc -> border + border-dir
  mergeRules(rule, properties, (rules, lastNode) => {
    if (rules.some(stylehacks.detect)) {
      return false;
    ***REMOVED***

    const values = rules.map((node) => parseTrbl(node.value));
    const mapped = [0, 1, 2, 3].map((i) =>
      [values[0][i], values[1][i], values[2][i]].join(' ')
    );

    if (!canMergeValues(mapped)) {
      return false;
    ***REMOVED***

    const [width, style, color] = rules;
    const reduced = getDistinctShorthands(mapped);

    if (isCloseEnough(mapped) && canMerge(rules, false)) {
      const first =
        mapped.indexOf(reduced[0]) !== mapped.lastIndexOf(reduced[0]);

      const border = insertCloned(
        /** @type {import('postcss').Rule***REMOVED*** */ (lastNode.parent),
        lastNode,
        {
          prop: 'border',
          value: first ? reduced[0] : reduced[1],
        ***REMOVED***
      );

      if (reduced[1]) {
        const value = first ? reduced[1] : reduced[0];
        const prop = borderProperty(trbl[mapped.indexOf(value)]);

        rule.insertAfter(
          border,
          Object.assign(lastNode.clone(), {
            prop,
            value,
          ***REMOVED***)
        );
      ***REMOVED***
      for (const node of rules) {
        node.remove();
      ***REMOVED***

      return true;
    ***REMOVED*** else if (reduced.length === 1) {
      rule.insertBefore(
        color,
        Object.assign(lastNode.clone(), {
          prop: 'border',
          value: [width, style].map(getValue).join(' '),
        ***REMOVED***)
      );
      rules
        .filter((node) => node.prop.toLowerCase() !== properties[2])
        .forEach((node) => node.remove());

      return true;
    ***REMOVED***
    return false;
  ***REMOVED***);

  // border-wsc -> border + border-trbl
  mergeRules(rule, properties, (rules, lastNode) => {
    if (rules.some(stylehacks.detect)) {
      return false;
    ***REMOVED***

    const values = rules.map((node) => parseTrbl(node.value));
    const mapped = [0, 1, 2, 3].map((i) =>
      [values[0][i], values[1][i], values[2][i]].join(' ')
    );
    const reduced = getDistinctShorthands(mapped);
    const none = 'medium none currentcolor';

    if (reduced.length > 1 && reduced.length < 4 && reduced.includes(none)) {
      const filtered = mapped.filter((p) => p !== none);
      const mostCommon = reduced.sort(
        (a, b) =>
          mapped.filter((v) => v === b).length -
          mapped.filter((v) => v === a).length
      )[0];
      const borderValue = reduced.length === 2 ? filtered[0] : mostCommon;

      rule.insertBefore(
        lastNode,
        Object.assign(lastNode.clone(), {
          prop: 'border',
          value: borderValue,
        ***REMOVED***)
      );

      directions.forEach((dir, i) => {
        if (mapped[i] !== borderValue) {
          rule.insertBefore(
            lastNode,
            Object.assign(lastNode.clone(), {
              prop: dir,
              value: mapped[i],
            ***REMOVED***)
          );
        ***REMOVED***
      ***REMOVED***);

      for (const node of rules) {
        node.remove();
      ***REMOVED***

      return true;
    ***REMOVED***
    return false;
  ***REMOVED***);

  // border-trbl -> border
  // border-trbl -> border + border-trbl
  mergeRules(rule, directions, (rules, lastNode) => {
    if (rules.some(stylehacks.detect)) {
      return false;
    ***REMOVED***

    const values = rules.map((node) => {
      const wscValue = parseWsc(node.value);

      if (!isValidWsc(wscValue)) {
        return node.value;
      ***REMOVED***

      return wscValue.map((value, i) => value || defaults[i]).join(' ');
    ***REMOVED***);

    const reduced = getDistinctShorthands(values);

    if (isCloseEnough(values)) {
      const first =
        values.indexOf(reduced[0]) !== values.lastIndexOf(reduced[0]);

      rule.insertBefore(
        lastNode,
        Object.assign(lastNode.clone(), {
          prop: 'border',
          value: minifyWsc(first ? values[0] : values[1]),
        ***REMOVED***)
      );

      if (reduced[1]) {
        const value = first ? reduced[1] : reduced[0];
        const prop = directions[values.indexOf(value)];
        rule.insertBefore(
          lastNode,
          Object.assign(lastNode.clone(), {
            prop: prop,
            value: minifyWsc(value),
          ***REMOVED***)
        );
      ***REMOVED***

      for (const node of rules) {
        node.remove();
      ***REMOVED***

      return true;
    ***REMOVED***
    return false;
  ***REMOVED***);

  // border-trbl-wsc + border-trbl (custom prop) -> border-trbl + border-trbl-wsc (custom prop)
  directions.forEach((direction) => {
    wsc.forEach((style, i) => {
      const prop = `${direction***REMOVED***-${style***REMOVED***`;

      mergeRules(rule, [direction, prop], (rules, lastNode) => {
        if (lastNode.prop !== direction) {
          return false;
        ***REMOVED***

        const values = parseWsc(lastNode.value);

        if (!isValidWsc(values)) {
          return false;
        ***REMOVED***

        const wscProp = rules.filter((r) => r !== lastNode)[0];

        if (!isValueCustomProp(values[i]) || isCustomProp(wscProp)) {
          return false;
        ***REMOVED***

        const wscValue = values[i];

        values[i] = wscProp.value;

        if (canMerge(rules, false) && !rules.some(stylehacks.detect)) {
          insertCloned(
            /** @type {import('postcss').Rule***REMOVED*** */ (lastNode.parent),
            lastNode,
            {
              prop,
              value: wscValue,
            ***REMOVED***
          );
          lastNode.value = minifyWsc(/** @type {any***REMOVED*** */ (values));

          wscProp.remove();

          return true;
        ***REMOVED***
        return false;
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***);

  // border-wsc + border (custom prop) -> border + border-wsc (custom prop)
  wsc.forEach((style, i) => {
    const prop = borderProperty(style);
    mergeRules(rule, ['border', prop], (rules, lastNode) => {
      if (lastNode.prop !== 'border') {
        return false;
      ***REMOVED***

      const values = parseWsc(lastNode.value);

      if (!isValidWsc(values)) {
        return false;
      ***REMOVED***

      const wscProp = rules.filter((r) => r !== lastNode)[0];

      if (!isValueCustomProp(values[i]) || isCustomProp(wscProp)) {
        return false;
      ***REMOVED***

      const wscValue = values[i];

      values[i] = wscProp.value;

      if (canMerge(rules, false) && !rules.some(stylehacks.detect)) {
        insertCloned(
          /** @type {import('postcss').Rule***REMOVED*** */ (lastNode.parent),
          lastNode,
          {
            prop,
            value: wscValue,
          ***REMOVED***
        );
        lastNode.value = minifyWsc(/** @type {any***REMOVED*** */ (values));
        wscProp.remove();

        return true;
      ***REMOVED***
      return false;
    ***REMOVED***);
  ***REMOVED***);

  // optimize border-trbl
  let decls = getDecls(rule, directions);

  while (decls.length) {
    const lastNode = decls[decls.length - 1];

    wsc.forEach((d, i) => {
      const names = directions
        .filter((name) => name !== lastNode.prop)
        .map((name) => `${name***REMOVED***-${d***REMOVED***`);

      let nodes = rule.nodes.slice(0, rule.nodes.indexOf(lastNode));

      const border = getLastNode(nodes, 'border');

      if (border) {
        nodes = nodes.slice(nodes.indexOf(border));
      ***REMOVED***

      const props = nodes.filter(
        (node) =>
          node.type === 'decl' &&
          names.includes(node.prop) &&
          node.important === lastNode.important
      );
      const rules = getRules(
        /** @type {import('postcss').Declaration[]***REMOVED*** */ (props),
        names
      );

      if (hasAllProps(rules, ...names) && !rules.some(stylehacks.detect)) {
        const values = rules.map((node) => (node ? node.value : null));
        const filteredValues = values.filter(Boolean);
        const lastNodeValue = list.space(lastNode.value)[i];

        values[directions.indexOf(lastNode.prop)] = lastNodeValue;

        let value = minifyTrbl(values.join(' '));

        if (
          filteredValues[0] === filteredValues[1] &&
          filteredValues[1] === filteredValues[2]
        ) {
          value = /** @type {string***REMOVED*** */ (filteredValues[0]);
        ***REMOVED***

        let refNode = props[props.length - 1];

        if (value === lastNodeValue) {
          refNode = lastNode;
          let valueArray = list.space(lastNode.value);
          valueArray.splice(i, 1);
          lastNode.value = valueArray.join(' ');
        ***REMOVED***

        insertCloned(
          /** @type {import('postcss').Rule***REMOVED*** */ (refNode.parent),
          /** @type {import('postcss').Declaration***REMOVED*** */ (refNode),
          {
            prop: borderProperty(d),
            value,
          ***REMOVED***
        );

        decls = decls.filter((node) => !rules.includes(node));
        for (const node of rules) {
          node.remove();
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***);

    decls = decls.filter((node) => node !== lastNode);
  ***REMOVED***

  rule.walkDecls('border', (decl) => {
    const nextDecl = decl.next();

    if (!nextDecl || nextDecl.type !== 'decl') {
      return false;
    ***REMOVED***

    const index = directions.indexOf(nextDecl.prop);

    if (index === -1) {
      return;
    ***REMOVED***

    const values = parseWsc(decl.value);
    const nextValues = parseWsc(nextDecl.value);

    if (!isValidWsc(values) || !isValidWsc(nextValues)) {
      return;
    ***REMOVED***

    const config = {
      values,
      nextValues,
      decl,
      nextDecl,
      index,
    ***REMOVED***;

    return mergeRedundant(config);
  ***REMOVED***);

  rule.walkDecls(/^border($|-(top|right|bottom|left)$)/i, (decl) => {
    let values = parseWsc(decl.value);

    if (!isValidWsc(values)) {
      return;
    ***REMOVED***

    const position = directions.indexOf(decl.prop);
    let dirs = [...directions];

    dirs.splice(position, 1);
    wsc.forEach((d, i) => {
      const props = dirs.map((dir) => `${dir***REMOVED***-${d***REMOVED***`);

      mergeRules(rule, [decl.prop, ...props], (rules) => {
        if (!rules.includes(decl)) {
          return false;
        ***REMOVED***

        const longhands = rules.filter((p) => p !== decl);

        if (
          longhands[0].value.toLowerCase() ===
            longhands[1].value.toLowerCase() &&
          longhands[1].value.toLowerCase() ===
            longhands[2].value.toLowerCase() &&
          values[i] !== undefined &&
          longhands[0].value.toLowerCase() === values[i].toLowerCase()
        ) {
          for (const node of longhands) {
            node.remove();
          ***REMOVED***

          insertCloned(
            /** @type {import('postcss').Rule***REMOVED*** */ (decl.parent),
            decl,
            {
              prop: borderProperty(d),
              value: values[i],
            ***REMOVED***
          );

          /** @type {string|null***REMOVED*** */ (values[i]) = null;
        ***REMOVED***
        return false;
      ***REMOVED***);

      const newValue = values.join(' ');

      if (newValue) {
        decl.value = newValue;
      ***REMOVED*** else {
        decl.remove();
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***);

  // clean-up values
  rule.walkDecls(/^border($|-(top|right|bottom|left)$)/i, (decl) => {
    decl.value = minifyWsc(decl.value);
  ***REMOVED***);

  // border-spacing-hv -> border-spacing
  rule.walkDecls(/^border-spacing$/i, (decl) => {
    const value = list.space(decl.value);

    // merge vertical and horizontal dups
    if (value.length > 1 && value[0] === value[1]) {
      decl.value = value.slice(1).join(' ');
    ***REMOVED***
  ***REMOVED***);

  // clean-up rules
  decls = getDecls(rule, allProperties);

  while (decls.length) {
    const lastNode = decls[decls.length - 1];
    const lastPart = lastNode.prop.split('-').pop();

    // remove properties of lower precedence
    const lesser = decls.filter(
      (node) =>
        !stylehacks.detect(lastNode) &&
        !stylehacks.detect(node) &&
        !isCustomProp(lastNode) &&
        node !== lastNode &&
        node.important === lastNode.important &&
        /** @type {number***REMOVED*** */ (getLevel(node.prop)) >
          /** @type {number***REMOVED*** */ (getLevel(lastNode.prop)) &&
        (node.prop.toLowerCase().includes(lastNode.prop) ||
          node.prop.toLowerCase().endsWith(/** @type {string***REMOVED*** */ (lastPart)))
    );

    for (const node of lesser) {
      node.remove();
    ***REMOVED***
    decls = decls.filter((node) => !lesser.includes(node));

    // get duplicate properties
    let duplicates = decls.filter(
      (node) =>
        !stylehacks.detect(lastNode) &&
        !stylehacks.detect(node) &&
        node !== lastNode &&
        node.important === lastNode.important &&
        node.prop === lastNode.prop &&
        !(!isCustomProp(node) && isCustomProp(lastNode))
    );

    if (duplicates.length) {
      if (colorMightRequireFallback.test(getColorValue(lastNode))) {
        const preserve = duplicates
          .filter(
            (node) => !colorMightRequireFallback.test(getColorValue(node))
          )
          .pop();

        duplicates = duplicates.filter((node) => node !== preserve);
      ***REMOVED***
      for (const node of duplicates) {
        node.remove();
      ***REMOVED***
    ***REMOVED***

    decls = decls.filter(
      (node) => node !== lastNode && !duplicates.includes(node)
    );
  ***REMOVED***
***REMOVED***

module.exports = {
  explode,
  merge,
***REMOVED***;
