// @nodelib/fs.scandir/out/providers/async.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.readdir = exports.readdirWithFileTypes = exports.read = void 0;
const fsStat = require("@nodelib/fs.stat");
const rpl = require("run-parallel");
const constants_1 = require("../constants");
const utils = require("../utils");
const common = require("./common");
function read(directory, settings, callback) {
    if (!settings.stats && constants_1.IS_SUPPORT_READDIR_WITH_FILE_TYPES) {
        readdirWithFileTypes(directory, settings, callback);
        return;
    ***REMOVED***
    readdir(directory, settings, callback);
***REMOVED***
exports.read = read;
function readdirWithFileTypes(directory, settings, callback) {
    settings.fs.readdir(directory, { withFileTypes: true ***REMOVED***, (readdirError, dirents) => {
        if (readdirError !== null) {
            callFailureCallback(callback, readdirError);
            return;
        ***REMOVED***
        const entries = dirents.map((dirent) => ({
            dirent,
            name: dirent.name,
            path: common.joinPathSegments(directory, dirent.name, settings.pathSegmentSeparator)
        ***REMOVED***));
        if (!settings.followSymbolicLinks) {
            callSuccessCallback(callback, entries);
            return;
        ***REMOVED***
        const tasks = entries.map((entry) => makeRplTaskEntry(entry, settings));
        rpl(tasks, (rplError, rplEntries) => {
            if (rplError !== null) {
                callFailureCallback(callback, rplError);
                return;
            ***REMOVED***
            callSuccessCallback(callback, rplEntries);
        ***REMOVED***);
    ***REMOVED***);
***REMOVED***
exports.readdirWithFileTypes = readdirWithFileTypes;
function makeRplTaskEntry(entry, settings) {
    return (done) => {
        if (!entry.dirent.isSymbolicLink()) {
            done(null, entry);
            return;
        ***REMOVED***
        settings.fs.stat(entry.path, (statError, stats) => {
            if (statError !== null) {
                if (settings.throwErrorOnBrokenSymbolicLink) {
                    done(statError);
                    return;
                ***REMOVED***
                done(null, entry);
                return;
            ***REMOVED***
            entry.dirent = utils.fs.createDirentFromStats(entry.name, stats);
            done(null, entry);
        ***REMOVED***);
    ***REMOVED***;
***REMOVED***
function readdir(directory, settings, callback) {
    settings.fs.readdir(directory, (readdirError, names) => {
        if (readdirError !== null) {
            callFailureCallback(callback, readdirError);
            return;
        ***REMOVED***
        const tasks = names.map((name) => {
            const path = common.joinPathSegments(directory, name, settings.pathSegmentSeparator);
            return (done) => {
                fsStat.stat(path, settings.fsStatSettings, (error, stats) => {
                    if (error !== null) {
                        done(error);
                        return;
                    ***REMOVED***
                    const entry = {
                        name,
                        path,
                        dirent: utils.fs.createDirentFromStats(name, stats)
                    ***REMOVED***;
                    if (settings.stats) {
                        entry.stats = stats;
                    ***REMOVED***
                    done(null, entry);
                ***REMOVED***);
            ***REMOVED***;
        ***REMOVED***);
        rpl(tasks, (rplError, entries) => {
            if (rplError !== null) {
                callFailureCallback(callback, rplError);
                return;
            ***REMOVED***
            callSuccessCallback(callback, entries);
        ***REMOVED***);
    ***REMOVED***);
***REMOVED***
exports.readdir = readdir;
function callFailureCallback(callback, error) {
    callback(error);
***REMOVED***
function callSuccessCallback(callback, result) {
    callback(null, result);
***REMOVED***
