// webpack-dev-server/lib/servers/SockJSServer.js
"use strict";

const sockjs = require("sockjs");
const BaseServer = require("./BaseServer");

/** @typedef {import("../Server").WebSocketServerConfiguration***REMOVED*** WebSocketServerConfiguration */
/** @typedef {import("../Server").ClientConnection***REMOVED*** ClientConnection */

// Workaround for sockjs@~0.3.19
// sockjs will remove Origin header, however Origin header is required for checking host.
// See https://github.com/webpack/webpack-dev-server/issues/1604 for more information
{
  // @ts-ignore
  const SockjsSession = require("sockjs/lib/transport").Session;
  const decorateConnection = SockjsSession.prototype.decorateConnection;

  /**
   * @param {import("http").IncomingMessage***REMOVED*** req
   */
  // eslint-disable-next-line func-names
  SockjsSession.prototype.decorateConnection = function (req) {
    decorateConnection.call(this, req);

    const connection = this.connection;

    if (
      connection.headers &&
      !("origin" in connection.headers) &&
      "origin" in req.headers
    ) {
      connection.headers.origin = req.headers.origin;
    ***REMOVED***
  ***REMOVED***;
***REMOVED***

module.exports = class SockJSServer extends BaseServer {
  // options has: error (function), debug (function), server (http/s server), path (string)
  /**
   * @param {import("../Server")***REMOVED*** server
   */
  constructor(server) {
    super(server);

    const webSocketServerOptions =
      /** @type {NonNullable<WebSocketServerConfiguration["options"]>***REMOVED*** */
      (
        /** @type {WebSocketServerConfiguration***REMOVED*** */
        (this.server.options.webSocketServer).options
      );

    /**
     * @param {NonNullable<WebSocketServerConfiguration["options"]>***REMOVED*** options
     * @returns {string***REMOVED***
     */
    const getSockjsUrl = (options) => {
      if (typeof options.sockjsUrl !== "undefined") {
        return options.sockjsUrl;
      ***REMOVED***

      return "/__webpack_dev_server__/sockjs.bundle.js";
    ***REMOVED***;

    this.implementation = sockjs.createServer({
      // Use provided up-to-date sockjs-client
      sockjs_url: getSockjsUrl(webSocketServerOptions),
      // Default logger is very annoy. Limit useless logs.
      /**
       * @param {string***REMOVED*** severity
       * @param {string***REMOVED*** line
       */
      log: (severity, line) => {
        if (severity === "error") {
          this.server.logger.error(line);
        ***REMOVED*** else if (severity === "info") {
          this.server.logger.log(line);
        ***REMOVED*** else {
          this.server.logger.debug(line);
        ***REMOVED***
      ***REMOVED***,
    ***REMOVED***);

    /**
     * @param {import("sockjs").ServerOptions & { path?: string ***REMOVED******REMOVED*** options
     * @returns {string | undefined***REMOVED***
     */
    const getPrefix = (options) => {
      if (typeof options.prefix !== "undefined") {
        return options.prefix;
      ***REMOVED***

      return options.path;
    ***REMOVED***;

    const options = {
      ...webSocketServerOptions,
      prefix: getPrefix(webSocketServerOptions),
    ***REMOVED***;

    this.implementation.installHandlers(
      /** @type {import("http").Server***REMOVED*** */ (this.server.server),
      options
    );

    this.implementation.on("connection", (client) => {
      // @ts-ignore
      // Implement the the same API as for `ws`
      client.send = client.write;
      // @ts-ignore
      client.terminate = client.close;

      this.clients.push(/** @type {ClientConnection***REMOVED*** */ (client));

      client.on("close", () => {
        this.clients.splice(
          this.clients.indexOf(/** @type {ClientConnection***REMOVED*** */ (client)),
          1
        );
      ***REMOVED***);
    ***REMOVED***);

    // @ts-ignore
    this.implementation.close = (callback) => {
      callback();
    ***REMOVED***;
  ***REMOVED***
***REMOVED***;
