// axios/lib/core/enhanceError.js
'use strict';

/**
 * Update an Error with the specified config, error code, and response.
 *
 * @param {Error***REMOVED*** error The error to update.
 * @param {Object***REMOVED*** config The config.
 * @param {string***REMOVED*** [code] The error code (for example, 'ECONNABORTED').
 * @param {Object***REMOVED*** [request] The request.
 * @param {Object***REMOVED*** [response] The response.
 * @returns {Error***REMOVED*** The error.
 */
module.exports = function enhanceError(error, config, code, request, response) {
  error.config = config;
  if (code) {
    error.code = code;
  ***REMOVED***

  error.request = request;
  error.response = response;
  error.isAxiosError = true;

  error.toJSON = function toJSON() {
    return {
      // Standard
      message: this.message,
      name: this.name,
      // Microsoft
      description: this.description,
      number: this.number,
      // Mozilla
      fileName: this.fileName,
      lineNumber: this.lineNumber,
      columnNumber: this.columnNumber,
      stack: this.stack,
      // Axios
      config: this.config,
      code: this.code
    ***REMOVED***;
  ***REMOVED***;
  return error;
***REMOVED***;
