// webpack/lib/cache/PackFileCacheStrategy.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const FileSystemInfo = require("../FileSystemInfo");
const ProgressPlugin = require("../ProgressPlugin");
const { formatSize ***REMOVED*** = require("../SizeFormatHelpers");
const SerializerMiddleware = require("../serialization/SerializerMiddleware");
const LazySet = require("../util/LazySet");
const makeSerializable = require("../util/makeSerializable");
const memoize = require("../util/memoize");
const {
	createFileSerializer,
	NOT_SERIALIZABLE
***REMOVED*** = require("../util/serialization");

/** @typedef {import("../../declarations/WebpackOptions").SnapshotOptions***REMOVED*** SnapshotOptions */
/** @typedef {import("../Cache").Data***REMOVED*** Data */
/** @typedef {import("../Cache").Etag***REMOVED*** Etag */
/** @typedef {import("../Compiler")***REMOVED*** Compiler */
/** @typedef {import("../FileSystemInfo").ResolveBuildDependenciesResult***REMOVED*** ResolveBuildDependenciesResult */
/** @typedef {import("../FileSystemInfo").ResolveResults***REMOVED*** ResolveResults */
/** @typedef {import("../FileSystemInfo").Snapshot***REMOVED*** Snapshot */
/** @typedef {import("../logging/Logger").Logger***REMOVED*** Logger */
/** @typedef {import("../serialization/ObjectMiddleware").ObjectDeserializerContext***REMOVED*** ObjectDeserializerContext */
/** @typedef {import("../serialization/ObjectMiddleware").ObjectSerializerContext***REMOVED*** ObjectSerializerContext */
/** @typedef {typeof import("../util/Hash")***REMOVED*** Hash */
/** @typedef {import("../util/fs").IntermediateFileSystem***REMOVED*** IntermediateFileSystem */

/** @typedef {Set<string>***REMOVED*** Items */
/** @typedef {Set<string>***REMOVED*** BuildDependencies */
/** @typedef {Map<string, PackItemInfo>***REMOVED*** ItemInfo */

class PackContainer {
	/**
	 * @param {Pack***REMOVED*** data stored data
	 * @param {string***REMOVED*** version version identifier
	 * @param {Snapshot***REMOVED*** buildSnapshot snapshot of all build dependencies
	 * @param {BuildDependencies***REMOVED*** buildDependencies list of all unresolved build dependencies captured
	 * @param {ResolveResults***REMOVED*** resolveResults result of the resolved build dependencies
	 * @param {Snapshot***REMOVED*** resolveBuildDependenciesSnapshot snapshot of the dependencies of the build dependencies resolving
	 */
	constructor(
		data,
		version,
		buildSnapshot,
		buildDependencies,
		resolveResults,
		resolveBuildDependenciesSnapshot
	) {
		this.data = data;
		this.version = version;
		this.buildSnapshot = buildSnapshot;
		this.buildDependencies = buildDependencies;
		this.resolveResults = resolveResults;
		this.resolveBuildDependenciesSnapshot = resolveBuildDependenciesSnapshot;
	***REMOVED***

	/**
	 * @param {ObjectSerializerContext***REMOVED*** context context
	 */
	serialize({ write, writeLazy ***REMOVED***) {
		write(this.version);
		write(this.buildSnapshot);
		write(this.buildDependencies);
		write(this.resolveResults);
		write(this.resolveBuildDependenciesSnapshot);
		/** @type {NonNullable<ObjectSerializerContext["writeLazy"]>***REMOVED*** */
		(writeLazy)(this.data);
	***REMOVED***

	/**
	 * @param {ObjectDeserializerContext***REMOVED*** context context
	 */
	deserialize({ read ***REMOVED***) {
		this.version = read();
		this.buildSnapshot = read();
		this.buildDependencies = read();
		this.resolveResults = read();
		this.resolveBuildDependenciesSnapshot = read();
		this.data = read();
	***REMOVED***
***REMOVED***

makeSerializable(
	PackContainer,
	"webpack/lib/cache/PackFileCacheStrategy",
	"PackContainer"
);

const MIN_CONTENT_SIZE = 1024 * 1024; // 1 MB
const CONTENT_COUNT_TO_MERGE = 10;
const MIN_ITEMS_IN_FRESH_PACK = 100;
const MAX_ITEMS_IN_FRESH_PACK = 50000;
const MAX_TIME_IN_FRESH_PACK = 1 * 60 * 1000; // 1 min

class PackItemInfo {
	/**
	 * @param {string***REMOVED*** identifier identifier of item
	 * @param {string | null | undefined***REMOVED*** etag etag of item
	 * @param {Data***REMOVED*** value fresh value of item
	 */
	constructor(identifier, etag, value) {
		this.identifier = identifier;
		this.etag = etag;
		this.location = -1;
		this.lastAccess = Date.now();
		this.freshValue = value;
	***REMOVED***
***REMOVED***

class Pack {
	/**
	 * @param {Logger***REMOVED*** logger a logger
	 * @param {number***REMOVED*** maxAge max age of cache items
	 */
	constructor(logger, maxAge) {
		/** @type {ItemInfo***REMOVED*** */
		this.itemInfo = new Map();
		/** @type {(string | undefined)[]***REMOVED*** */
		this.requests = [];
		this.requestsTimeout = undefined;
		/** @type {ItemInfo***REMOVED*** */
		this.freshContent = new Map();
		/** @type {(undefined | PackContent)[]***REMOVED*** */
		this.content = [];
		this.invalid = false;
		this.logger = logger;
		this.maxAge = maxAge;
	***REMOVED***

	/**
	 * @param {string***REMOVED*** identifier identifier
	 */
	_addRequest(identifier) {
		this.requests.push(identifier);
		if (this.requestsTimeout === undefined) {
			this.requestsTimeout = setTimeout(() => {
				this.requests.push(undefined);
				this.requestsTimeout = undefined;
			***REMOVED***, MAX_TIME_IN_FRESH_PACK);
			if (this.requestsTimeout.unref) this.requestsTimeout.unref();
		***REMOVED***
	***REMOVED***

	stopCapturingRequests() {
		if (this.requestsTimeout !== undefined) {
			clearTimeout(this.requestsTimeout);
			this.requestsTimeout = undefined;
		***REMOVED***
	***REMOVED***

	/**
	 * @param {string***REMOVED*** identifier unique name for the resource
	 * @param {string | null***REMOVED*** etag etag of the resource
	 * @returns {Data***REMOVED*** cached content
	 */
	get(identifier, etag) {
		const info = this.itemInfo.get(identifier);
		this._addRequest(identifier);
		if (info === undefined) {
			return;
		***REMOVED***
		if (info.etag !== etag) return null;
		info.lastAccess = Date.now();
		const loc = info.location;
		if (loc === -1) {
			return info.freshValue;
		***REMOVED***
		if (!this.content[loc]) {
			return;
		***REMOVED***
		return /** @type {PackContent***REMOVED*** */ (this.content[loc]).get(identifier);
	***REMOVED***

	/**
	 * @param {string***REMOVED*** identifier unique name for the resource
	 * @param {string | null***REMOVED*** etag etag of the resource
	 * @param {Data***REMOVED*** data cached content
	 * @returns {void***REMOVED***
	 */
	set(identifier, etag, data) {
		if (!this.invalid) {
			this.invalid = true;
			this.logger.log(`Pack got invalid because of write to: ${identifier***REMOVED***`);
		***REMOVED***
		const info = this.itemInfo.get(identifier);
		if (info === undefined) {
			const newInfo = new PackItemInfo(identifier, etag, data);
			this.itemInfo.set(identifier, newInfo);
			this._addRequest(identifier);
			this.freshContent.set(identifier, newInfo);
		***REMOVED*** else {
			const loc = info.location;
			if (loc >= 0) {
				this._addRequest(identifier);
				this.freshContent.set(identifier, info);
				const content = /** @type {PackContent***REMOVED*** */ (this.content[loc]);
				content.delete(identifier);
				if (content.items.size === 0) {
					this.content[loc] = undefined;
					this.logger.debug("Pack %d got empty and is removed", loc);
				***REMOVED***
			***REMOVED***
			info.freshValue = data;
			info.lastAccess = Date.now();
			info.etag = etag;
			info.location = -1;
		***REMOVED***
	***REMOVED***

	getContentStats() {
		let count = 0;
		let size = 0;
		for (const content of this.content) {
			if (content !== undefined) {
				count++;
				const s = content.getSize();
				if (s > 0) {
					size += s;
				***REMOVED***
			***REMOVED***
		***REMOVED***
		return { count, size ***REMOVED***;
	***REMOVED***

	/**
	 * @returns {number***REMOVED*** new location of data entries
	 */
	_findLocation() {
		let i;
		for (i = 0; i < this.content.length && this.content[i] !== undefined; i++);
		return i;
	***REMOVED***

	/**
	 * @private
	 * @param {Items***REMOVED*** items items
	 * @param {Items***REMOVED*** usedItems used items
	 * @param {number***REMOVED*** newLoc new location
	 */
	_gcAndUpdateLocation(items, usedItems, newLoc) {
		let count = 0;
		let lastGC;
		const now = Date.now();
		for (const identifier of items) {
			const info = /** @type {PackItemInfo***REMOVED*** */ (this.itemInfo.get(identifier));
			if (now - info.lastAccess > this.maxAge) {
				this.itemInfo.delete(identifier);
				items.delete(identifier);
				usedItems.delete(identifier);
				count++;
				lastGC = identifier;
			***REMOVED*** else {
				info.location = newLoc;
			***REMOVED***
		***REMOVED***
		if (count > 0) {
			this.logger.log(
				"Garbage Collected %d old items at pack %d (%d items remaining) e. g. %s",
				count,
				newLoc,
				items.size,
				lastGC
			);
		***REMOVED***
	***REMOVED***

	_persistFreshContent() {
		/** @typedef {{ items: Items, map: Content, loc: number ***REMOVED******REMOVED*** PackItem */
		const itemsCount = this.freshContent.size;
		if (itemsCount > 0) {
			const packCount = Math.ceil(itemsCount / MAX_ITEMS_IN_FRESH_PACK);
			const itemsPerPack = Math.ceil(itemsCount / packCount);
			/** @type {PackItem[]***REMOVED*** */
			const packs = [];
			let i = 0;
			let ignoreNextTimeTick = false;
			const createNextPack = () => {
				const loc = this._findLocation();
				this.content[loc] = /** @type {EXPECTED_ANY***REMOVED*** */ (null); // reserve
				/** @type {PackItem***REMOVED*** */
				const pack = {
					items: new Set(),
					map: new Map(),
					loc
				***REMOVED***;
				packs.push(pack);
				return pack;
			***REMOVED***;
			let pack = createNextPack();
			if (this.requestsTimeout !== undefined)
				clearTimeout(this.requestsTimeout);
			for (const identifier of this.requests) {
				if (identifier === undefined) {
					if (ignoreNextTimeTick) {
						ignoreNextTimeTick = false;
					***REMOVED*** else if (pack.items.size >= MIN_ITEMS_IN_FRESH_PACK) {
						i = 0;
						pack = createNextPack();
					***REMOVED***
					continue;
				***REMOVED***
				const info = this.freshContent.get(identifier);
				if (info === undefined) continue;
				pack.items.add(identifier);
				pack.map.set(identifier, info.freshValue);
				info.location = pack.loc;
				info.freshValue = undefined;
				this.freshContent.delete(identifier);
				if (++i > itemsPerPack) {
					i = 0;
					pack = createNextPack();
					ignoreNextTimeTick = true;
				***REMOVED***
			***REMOVED***
			this.requests.length = 0;
			for (const pack of packs) {
				this.content[pack.loc] = new PackContent(
					pack.items,
					new Set(pack.items),
					new PackContentItems(pack.map)
				);
			***REMOVED***
			this.logger.log(
				`${itemsCount***REMOVED*** fresh items in cache put into pack ${
					packs.length > 1
						? packs
								.map(pack => `${pack.loc***REMOVED*** (${pack.items.size***REMOVED*** items)`)
								.join(", ")
						: packs[0].loc
				***REMOVED***`
			);
		***REMOVED***
	***REMOVED***

	/**
	 * Merges small content files to a single content file
	 */
	_optimizeSmallContent() {
		// 1. Find all small content files
		// Treat unused content files separately to avoid
		// a merge-split cycle
		/** @type {number[]***REMOVED*** */
		const smallUsedContents = [];
		/** @type {number***REMOVED*** */
		let smallUsedContentSize = 0;
		/** @type {number[]***REMOVED*** */
		const smallUnusedContents = [];
		/** @type {number***REMOVED*** */
		let smallUnusedContentSize = 0;
		for (let i = 0; i < this.content.length; i++) {
			const content = this.content[i];
			if (content === undefined) continue;
			if (content.outdated) continue;
			const size = content.getSize();
			if (size < 0 || size > MIN_CONTENT_SIZE) continue;
			if (content.used.size > 0) {
				smallUsedContents.push(i);
				smallUsedContentSize += size;
			***REMOVED*** else {
				smallUnusedContents.push(i);
				smallUnusedContentSize += size;
			***REMOVED***
		***REMOVED***

		// 2. Check if minimum number is reached
		let mergedIndices;
		if (
			smallUsedContents.length >= CONTENT_COUNT_TO_MERGE ||
			smallUsedContentSize > MIN_CONTENT_SIZE
		) {
			mergedIndices = smallUsedContents;
		***REMOVED*** else if (
			smallUnusedContents.length >= CONTENT_COUNT_TO_MERGE ||
			smallUnusedContentSize > MIN_CONTENT_SIZE
		) {
			mergedIndices = smallUnusedContents;
		***REMOVED*** else return;

		/** @type {PackContent[] ***REMOVED*** */
		const mergedContent = [];

		// 3. Remove old content entries
		for (const i of mergedIndices) {
			mergedContent.push(/** @type {PackContent***REMOVED*** */ (this.content[i]));
			this.content[i] = undefined;
		***REMOVED***

		// 4. Determine merged items
		/** @type {Items***REMOVED*** */
		const mergedItems = new Set();
		/** @type {Items***REMOVED*** */
		const mergedUsedItems = new Set();
		/** @type {((map: Content) => Promise<void>)[]***REMOVED*** */
		const addToMergedMap = [];
		for (const content of mergedContent) {
			for (const identifier of content.items) {
				mergedItems.add(identifier);
			***REMOVED***
			for (const identifier of content.used) {
				mergedUsedItems.add(identifier);
			***REMOVED***
			addToMergedMap.push(async map => {
				// unpack existing content
				// after that values are accessible in .content
				await content.unpack(
					"it should be merged with other small pack contents"
				);
				for (const [identifier, value] of /** @type {Content***REMOVED*** */ (
					content.content
				)) {
					map.set(identifier, value);
				***REMOVED***
			***REMOVED***);
		***REMOVED***

		// 5. GC and update location of merged items
		const newLoc = this._findLocation();
		this._gcAndUpdateLocation(mergedItems, mergedUsedItems, newLoc);

		// 6. If not empty, store content somewhere
		if (mergedItems.size > 0) {
			this.content[newLoc] = new PackContent(
				mergedItems,
				mergedUsedItems,
				memoize(async () => {
					/** @type {Content***REMOVED*** */
					const map = new Map();
					await Promise.all(addToMergedMap.map(fn => fn(map)));
					return new PackContentItems(map);
				***REMOVED***)
			);
			this.logger.log(
				"Merged %d small files with %d cache items into pack %d",
				mergedContent.length,
				mergedItems.size,
				newLoc
			);
		***REMOVED***
	***REMOVED***

	/**
	 * Split large content files with used and unused items
	 * into two parts to separate used from unused items
	 */
	_optimizeUnusedContent() {
		// 1. Find a large content file with used and unused items
		for (let i = 0; i < this.content.length; i++) {
			const content = this.content[i];
			if (content === undefined) continue;
			const size = content.getSize();
			if (size < MIN_CONTENT_SIZE) continue;
			const used = content.used.size;
			const total = content.items.size;
			if (used > 0 && used < total) {
				// 2. Remove this content
				this.content[i] = undefined;

				// 3. Determine items for the used content file
				const usedItems = new Set(content.used);
				const newLoc = this._findLocation();
				this._gcAndUpdateLocation(usedItems, usedItems, newLoc);

				// 4. Create content file for used items
				if (usedItems.size > 0) {
					this.content[newLoc] = new PackContent(
						usedItems,
						new Set(usedItems),
						async () => {
							await content.unpack(
								"it should be splitted into used and unused items"
							);
							/** @type {Content***REMOVED*** */
							const map = new Map();
							for (const identifier of usedItems) {
								map.set(
									identifier,
									/** @type {Content***REMOVED*** */
									(content.content).get(identifier)
								);
							***REMOVED***
							return new PackContentItems(map);
						***REMOVED***
					);
				***REMOVED***

				// 5. Determine items for the unused content file
				const unusedItems = new Set(content.items);
				const usedOfUnusedItems = new Set();
				for (const identifier of usedItems) {
					unusedItems.delete(identifier);
				***REMOVED***
				const newUnusedLoc = this._findLocation();
				this._gcAndUpdateLocation(unusedItems, usedOfUnusedItems, newUnusedLoc);

				// 6. Create content file for unused items
				if (unusedItems.size > 0) {
					this.content[newUnusedLoc] = new PackContent(
						unusedItems,
						usedOfUnusedItems,
						async () => {
							await content.unpack(
								"it should be splitted into used and unused items"
							);
							const map = new Map();
							for (const identifier of unusedItems) {
								map.set(
									identifier,
									/** @type {Content***REMOVED*** */
									(content.content).get(identifier)
								);
							***REMOVED***
							return new PackContentItems(map);
						***REMOVED***
					);
				***REMOVED***

				this.logger.log(
					"Split pack %d into pack %d with %d used items and pack %d with %d unused items",
					i,
					newLoc,
					usedItems.size,
					newUnusedLoc,
					unusedItems.size
				);

				// optimizing only one of them is good enough and
				// reduces the amount of serialization needed
				return;
			***REMOVED***
		***REMOVED***
	***REMOVED***

	/**
	 * Find the content with the oldest item and run GC on that.
	 * Only runs for one content to avoid large invalidation.
	 */
	_gcOldestContent() {
		/** @type {PackItemInfo | undefined***REMOVED*** */
		let oldest;
		for (const info of this.itemInfo.values()) {
			if (oldest === undefined || info.lastAccess < oldest.lastAccess) {
				oldest = info;
			***REMOVED***
		***REMOVED***
		if (
			Date.now() - /** @type {PackItemInfo***REMOVED*** */ (oldest).lastAccess >
			this.maxAge
		) {
			const loc = /** @type {PackItemInfo***REMOVED*** */ (oldest).location;
			if (loc < 0) return;
			const content = /** @type {PackContent***REMOVED*** */ (this.content[loc]);
			const items = new Set(content.items);
			const usedItems = new Set(content.used);
			this._gcAndUpdateLocation(items, usedItems, loc);

			this.content[loc] =
				items.size > 0
					? new PackContent(items, usedItems, async () => {
							await content.unpack(
								"it contains old items that should be garbage collected"
							);
							const map = new Map();
							for (const identifier of items) {
								map.set(
									identifier,
									/** @type {Content***REMOVED*** */
									(content.content).get(identifier)
								);
							***REMOVED***
							return new PackContentItems(map);
						***REMOVED***)
					: undefined;
		***REMOVED***
	***REMOVED***

	/**
	 * @param {ObjectSerializerContext***REMOVED*** context context
	 */
	serialize({ write, writeSeparate ***REMOVED***) {
		this._persistFreshContent();
		this._optimizeSmallContent();
		this._optimizeUnusedContent();
		this._gcOldestContent();
		for (const identifier of this.itemInfo.keys()) {
			write(identifier);
		***REMOVED***
		write(null); // null as marker of the end of keys
		for (const info of this.itemInfo.values()) {
			write(info.etag);
		***REMOVED***
		for (const info of this.itemInfo.values()) {
			write(info.lastAccess);
		***REMOVED***
		for (let i = 0; i < this.content.length; i++) {
			const content = this.content[i];
			if (content !== undefined) {
				write(content.items);
				content.writeLazy(lazy =>
					/** @type {NonNullable<ObjectSerializerContext["writeSeparate"]>***REMOVED*** */
					(writeSeparate)(lazy, { name: `${i***REMOVED***` ***REMOVED***)
				);
			***REMOVED*** else {
				write(undefined); // undefined marks an empty content slot
			***REMOVED***
		***REMOVED***
		write(null); // null as marker of the end of items
	***REMOVED***

	/**
	 * @param {ObjectDeserializerContext & { logger: Logger ***REMOVED******REMOVED*** context context
	 */
	deserialize({ read, logger ***REMOVED***) {
		this.logger = logger;
		{
			const items = [];
			let item = read();
			while (item !== null) {
				items.push(item);
				item = read();
			***REMOVED***
			this.itemInfo.clear();
			const infoItems = items.map(identifier => {
				const info = new PackItemInfo(identifier, undefined, undefined);
				this.itemInfo.set(identifier, info);
				return info;
			***REMOVED***);
			for (const info of infoItems) {
				info.etag = read();
			***REMOVED***
			for (const info of infoItems) {
				info.lastAccess = read();
			***REMOVED***
		***REMOVED***
		this.content.length = 0;
		let items = read();
		while (items !== null) {
			if (items === undefined) {
				this.content.push(items);
			***REMOVED*** else {
				const idx = this.content.length;
				const lazy = read();
				this.content.push(
					new PackContent(
						items,
						new Set(),
						lazy,
						logger,
						`${this.content.length***REMOVED***`
					)
				);
				for (const identifier of items) {
					/** @type {PackItemInfo***REMOVED*** */
					(this.itemInfo.get(identifier)).location = idx;
				***REMOVED***
			***REMOVED***
			items = read();
		***REMOVED***
	***REMOVED***
***REMOVED***

makeSerializable(Pack, "webpack/lib/cache/PackFileCacheStrategy", "Pack");

/** @typedef {Map<string, Data>***REMOVED*** Content */

class PackContentItems {
	/**
	 * @param {Content***REMOVED*** map items
	 */
	constructor(map) {
		this.map = map;
	***REMOVED***

	/**
	 * @param {ObjectSerializerContext & { logger: Logger, profile: boolean | undefined  ***REMOVED******REMOVED*** context context
	 */
	serialize({ write, snapshot, rollback, logger, profile ***REMOVED***) {
		if (profile) {
			write(false);
			for (const [key, value] of this.map) {
				const s = snapshot();
				try {
					write(key);
					const start = process.hrtime();
					write(value);
					const durationHr = process.hrtime(start);
					const duration = durationHr[0] * 1000 + durationHr[1] / 1e6;
					if (duration > 1) {
						if (duration > 500)
							logger.error(`Serialization of '${key***REMOVED***': ${duration***REMOVED*** ms`);
						else if (duration > 50)
							logger.warn(`Serialization of '${key***REMOVED***': ${duration***REMOVED*** ms`);
						else if (duration > 10)
							logger.info(`Serialization of '${key***REMOVED***': ${duration***REMOVED*** ms`);
						else if (duration > 5)
							logger.log(`Serialization of '${key***REMOVED***': ${duration***REMOVED*** ms`);
						else logger.debug(`Serialization of '${key***REMOVED***': ${duration***REMOVED*** ms`);
					***REMOVED***
				***REMOVED*** catch (err) {
					rollback(s);
					if (err === NOT_SERIALIZABLE) continue;
					const msg = "Skipped not serializable cache item";
					const notSerializableErr = /** @type {Error***REMOVED*** */ (err);
					if (notSerializableErr.message.includes("ModuleBuildError")) {
						logger.log(
							`${msg***REMOVED*** (in build error): ${notSerializableErr.message***REMOVED***`
						);
						logger.debug(
							`${msg***REMOVED*** '${key***REMOVED***' (in build error): ${notSerializableErr.stack***REMOVED***`
						);
					***REMOVED*** else {
						logger.warn(`${msg***REMOVED***: ${notSerializableErr.message***REMOVED***`);
						logger.debug(`${msg***REMOVED*** '${key***REMOVED***': ${notSerializableErr.stack***REMOVED***`);
					***REMOVED***
				***REMOVED***
			***REMOVED***
			write(null);
			return;
		***REMOVED***
		// Try to serialize all at once
		const s = snapshot();
		try {
			write(true);
			write(this.map);
		***REMOVED*** catch (_err) {
			rollback(s);

			// Try to serialize each item on it's own
			write(false);
			for (const [key, value] of this.map) {
				const s = snapshot();
				try {
					write(key);
					write(value);
				***REMOVED*** catch (err) {
					rollback(s);
					if (err === NOT_SERIALIZABLE) continue;
					const notSerializableErr = /** @type {Error***REMOVED*** */ (err);
					logger.warn(
						`Skipped not serializable cache item '${key***REMOVED***': ${notSerializableErr.message***REMOVED***`
					);
					logger.debug(notSerializableErr.stack);
				***REMOVED***
			***REMOVED***
			write(null);
		***REMOVED***
	***REMOVED***

	/**
	 * @param {ObjectDeserializerContext & { logger: Logger, profile: boolean | undefined ***REMOVED******REMOVED*** context context
	 */
	deserialize({ read, logger, profile ***REMOVED***) {
		if (read()) {
			this.map = read();
		***REMOVED*** else if (profile) {
			const map = new Map();
			let key = read();
			while (key !== null) {
				const start = process.hrtime();
				const value = read();
				const durationHr = process.hrtime(start);
				const duration = durationHr[0] * 1000 + durationHr[1] / 1e6;
				if (duration > 1) {
					if (duration > 100)
						logger.error(`Deserialization of '${key***REMOVED***': ${duration***REMOVED*** ms`);
					else if (duration > 20)
						logger.warn(`Deserialization of '${key***REMOVED***': ${duration***REMOVED*** ms`);
					else if (duration > 5)
						logger.info(`Deserialization of '${key***REMOVED***': ${duration***REMOVED*** ms`);
					else if (duration > 2)
						logger.log(`Deserialization of '${key***REMOVED***': ${duration***REMOVED*** ms`);
					else logger.debug(`Deserialization of '${key***REMOVED***': ${duration***REMOVED*** ms`);
				***REMOVED***
				map.set(key, value);
				key = read();
			***REMOVED***
			this.map = map;
		***REMOVED*** else {
			const map = new Map();
			let key = read();
			while (key !== null) {
				map.set(key, read());
				key = read();
			***REMOVED***
			this.map = map;
		***REMOVED***
	***REMOVED***
***REMOVED***

makeSerializable(
	PackContentItems,
	"webpack/lib/cache/PackFileCacheStrategy",
	"PackContentItems"
);

/** @typedef {(() => Promise<PackContentItems> | PackContentItems) & Partial<{ options: { size?: number ***REMOVED******REMOVED***>***REMOVED*** LazyFunction */

class PackContent {
	/*
		This class can be in these states:
		   |   this.lazy    | this.content | this.outdated | state
		A1 |   undefined    |     Map      |     false     | fresh content
		A2 |   undefined    |     Map      |     true      | (will not happen)
		B1 | lazy () => {***REMOVED***  |  undefined   |     false     | not deserialized
		B2 | lazy () => {***REMOVED***  |  undefined   |     true      | not deserialized, but some items has been removed
		C1 | lazy* () => {***REMOVED*** |     Map      |     false     | deserialized
		C2 | lazy* () => {***REMOVED*** |     Map      |     true      | deserialized, and some items has been removed

		this.used is a subset of this.items.
		this.items is a subset of this.content.keys() resp. this.lazy().map.keys()
		When this.outdated === false, this.items === this.content.keys() resp. this.lazy().map.keys()
		When this.outdated === true, this.items should be used to recreated this.lazy/this.content.
		When this.lazy and this.content is set, they contain the same data.
		this.get must only be called with a valid item from this.items.
		In state C this.lazy is unMemoized
	*/

	/**
	 * @param {Items***REMOVED*** items keys
	 * @param {Items***REMOVED*** usedItems used keys
	 * @param {PackContentItems | (() => Promise<PackContentItems>)***REMOVED*** dataOrFn sync or async content
	 * @param {Logger=***REMOVED*** logger logger for logging
	 * @param {string=***REMOVED*** lazyName name of dataOrFn for logging
	 */
	constructor(items, usedItems, dataOrFn, logger, lazyName) {
		this.items = items;
		/** @type {LazyFunction | undefined***REMOVED*** */
		this.lazy = typeof dataOrFn === "function" ? dataOrFn : undefined;
		/** @type {Content | undefined***REMOVED*** */
		this.content = typeof dataOrFn === "function" ? undefined : dataOrFn.map;
		this.outdated = false;
		this.used = usedItems;
		this.logger = logger;
		this.lazyName = lazyName;
	***REMOVED***

	/**
	 * @param {string***REMOVED*** identifier identifier
	 * @returns {string | Promise<string>***REMOVED*** result
	 */
	get(identifier) {
		this.used.add(identifier);
		if (this.content) {
			return this.content.get(identifier);
		***REMOVED***

		const logger = /** @type {Logger***REMOVED*** */ (this.logger);
		// We are in state B
		const { lazyName ***REMOVED*** = this;
		/** @type {string | undefined***REMOVED*** */
		let timeMessage;
		if (lazyName) {
			// only log once
			this.lazyName = undefined;
			timeMessage = `restore cache content ${lazyName***REMOVED*** (${formatSize(
				this.getSize()
			)***REMOVED***)`;
			logger.log(
				`starting to restore cache content ${lazyName***REMOVED*** (${formatSize(
					this.getSize()
				)***REMOVED***) because of request to: ${identifier***REMOVED***`
			);
			logger.time(timeMessage);
		***REMOVED***
		const value = /** @type {LazyFunction***REMOVED*** */ (this.lazy)();
		if ("then" in value) {
			return value.then(data => {
				const map = data.map;
				if (timeMessage) {
					logger.timeEnd(timeMessage);
				***REMOVED***
				// Move to state C
				this.content = map;
				this.lazy = SerializerMiddleware.unMemoizeLazy(this.lazy);
				return map.get(identifier);
			***REMOVED***);
		***REMOVED***

		const map = value.map;
		if (timeMessage) {
			logger.timeEnd(timeMessage);
		***REMOVED***
		// Move to state C
		this.content = map;
		this.lazy = SerializerMiddleware.unMemoizeLazy(this.lazy);
		return map.get(identifier);
	***REMOVED***

	/**
	 * @param {string***REMOVED*** reason explanation why unpack is necessary
	 * @returns {void | Promise<void>***REMOVED*** maybe a promise if lazy
	 */
	unpack(reason) {
		if (this.content) return;

		const logger = /** @type {Logger***REMOVED*** */ (this.logger);
		// Move from state B to C
		if (this.lazy) {
			const { lazyName ***REMOVED*** = this;
			/** @type {string | undefined***REMOVED*** */
			let timeMessage;
			if (lazyName) {
				// only log once
				this.lazyName = undefined;
				timeMessage = `unpack cache content ${lazyName***REMOVED*** (${formatSize(
					this.getSize()
				)***REMOVED***)`;
				logger.log(
					`starting to unpack cache content ${lazyName***REMOVED*** (${formatSize(
						this.getSize()
					)***REMOVED***) because ${reason***REMOVED***`
				);
				logger.time(timeMessage);
			***REMOVED***
			const value =
				/** @type {PackContentItems | Promise<PackContentItems>***REMOVED*** */
				(this.lazy());
			if ("then" in value) {
				return value.then(data => {
					if (timeMessage) {
						logger.timeEnd(timeMessage);
					***REMOVED***
					this.content = data.map;
				***REMOVED***);
			***REMOVED***
			if (timeMessage) {
				logger.timeEnd(timeMessage);
			***REMOVED***
			this.content = value.map;
		***REMOVED***
	***REMOVED***

	/**
	 * @returns {number***REMOVED*** size of the content or -1 if not known
	 */
	getSize() {
		if (!this.lazy) return -1;
		const options =
			/** @type {{ options: { size?: number ***REMOVED*** ***REMOVED******REMOVED*** */
			(this.lazy).options;
		if (!options) return -1;
		const size = options.size;
		if (typeof size !== "number") return -1;
		return size;
	***REMOVED***

	/**
	 * @param {string***REMOVED*** identifier identifier
	 */
	delete(identifier) {
		this.items.delete(identifier);
		this.used.delete(identifier);
		this.outdated = true;
	***REMOVED***

	/**
	 * @param {(lazy: LazyFunction) => (() => PackContentItems | Promise<PackContentItems>)***REMOVED*** write write function
	 * @returns {void***REMOVED***
	 */
	writeLazy(write) {
		if (!this.outdated && this.lazy) {
			// State B1 or C1
			// this.lazy is still the valid deserialized version
			write(this.lazy);
			return;
		***REMOVED***
		if (!this.outdated && this.content) {
			// State A1
			const map = new Map(this.content);
			// Move to state C1
			this.lazy = SerializerMiddleware.unMemoizeLazy(
				write(() => new PackContentItems(map))
			);
			return;
		***REMOVED***
		if (this.content) {
			// State A2 or C2
			/** @type {Content***REMOVED*** */
			const map = new Map();
			for (const item of this.items) {
				map.set(item, this.content.get(item));
			***REMOVED***
			// Move to state C1
			this.outdated = false;
			this.content = map;
			this.lazy = SerializerMiddleware.unMemoizeLazy(
				write(() => new PackContentItems(map))
			);
			return;
		***REMOVED***
		const logger = /** @type {Logger***REMOVED*** */ (this.logger);
		// State B2
		const { lazyName ***REMOVED*** = this;
		/** @type {string | undefined***REMOVED*** */
		let timeMessage;
		if (lazyName) {
			// only log once
			this.lazyName = undefined;
			timeMessage = `unpack cache content ${lazyName***REMOVED*** (${formatSize(
				this.getSize()
			)***REMOVED***)`;
			logger.log(
				`starting to unpack cache content ${lazyName***REMOVED*** (${formatSize(
					this.getSize()
				)***REMOVED***) because it's outdated and need to be serialized`
			);
			logger.time(timeMessage);
		***REMOVED***
		const value = /** @type {LazyFunction***REMOVED*** */ (this.lazy)();
		this.outdated = false;
		if ("then" in value) {
			// Move to state B1
			this.lazy = write(() =>
				value.then(data => {
					if (timeMessage) {
						logger.timeEnd(timeMessage);
					***REMOVED***
					const oldMap = data.map;
					/** @type {Content***REMOVED*** */
					const map = new Map();
					for (const item of this.items) {
						map.set(item, oldMap.get(item));
					***REMOVED***
					// Move to state C1 (or maybe C2)
					this.content = map;
					this.lazy = SerializerMiddleware.unMemoizeLazy(this.lazy);

					return new PackContentItems(map);
				***REMOVED***)
			);
		***REMOVED*** else {
			// Move to state C1
			if (timeMessage) {
				logger.timeEnd(timeMessage);
			***REMOVED***
			const oldMap = value.map;
			/** @type {Content***REMOVED*** */
			const map = new Map();
			for (const item of this.items) {
				map.set(item, oldMap.get(item));
			***REMOVED***
			this.content = map;
			this.lazy = write(() => new PackContentItems(map));
		***REMOVED***
	***REMOVED***
***REMOVED***

/**
 * @param {Buffer***REMOVED*** buf buffer
 * @returns {Buffer***REMOVED*** buffer that can be collected
 */
const allowCollectingMemory = buf => {
	const wasted = buf.buffer.byteLength - buf.byteLength;
	if (wasted > 8192 && (wasted > 1048576 || wasted > buf.byteLength)) {
		return Buffer.from(buf);
	***REMOVED***
	return buf;
***REMOVED***;

class PackFileCacheStrategy {
	/**
	 * @param {object***REMOVED*** options options
	 * @param {Compiler***REMOVED*** options.compiler the compiler
	 * @param {IntermediateFileSystem***REMOVED*** options.fs the filesystem
	 * @param {string***REMOVED*** options.context the context directory
	 * @param {string***REMOVED*** options.cacheLocation the location of the cache data
	 * @param {string***REMOVED*** options.version version identifier
	 * @param {Logger***REMOVED*** options.logger a logger
	 * @param {SnapshotOptions***REMOVED*** options.snapshot options regarding snapshotting
	 * @param {number***REMOVED*** options.maxAge max age of cache items
	 * @param {boolean | undefined***REMOVED*** options.profile track and log detailed timing information for individual cache items
	 * @param {boolean | undefined***REMOVED*** options.allowCollectingMemory allow to collect unused memory created during deserialization
	 * @param {false | "gzip" | "brotli" | undefined***REMOVED*** options.compression compression used
	 * @param {boolean | undefined***REMOVED*** options.readonly disable storing cache into filesystem
	 */
	constructor({
		compiler,
		fs,
		context,
		cacheLocation,
		version,
		logger,
		snapshot,
		maxAge,
		profile,
		allowCollectingMemory,
		compression,
		readonly
	***REMOVED***) {
		/** @type {import("../serialization/Serializer")<PackContainer, null, {***REMOVED***>***REMOVED*** */
		this.fileSerializer = createFileSerializer(
			fs,
			/** @type {string | Hash***REMOVED*** */
			(compiler.options.output.hashFunction)
		);
		this.fileSystemInfo = new FileSystemInfo(fs, {
			managedPaths: snapshot.managedPaths,
			immutablePaths: snapshot.immutablePaths,
			logger: logger.getChildLogger("webpack.FileSystemInfo"),
			hashFunction: compiler.options.output.hashFunction
		***REMOVED***);
		this.compiler = compiler;
		this.context = context;
		this.cacheLocation = cacheLocation;
		this.version = version;
		this.logger = logger;
		this.maxAge = maxAge;
		this.profile = profile;
		this.readonly = readonly;
		this.allowCollectingMemory = allowCollectingMemory;
		this.compression = compression;
		this._extension =
			compression === "brotli"
				? ".pack.br"
				: compression === "gzip"
					? ".pack.gz"
					: ".pack";
		this.snapshot = snapshot;
		/** @type {BuildDependencies***REMOVED*** */
		this.buildDependencies = new Set();
		/** @type {LazySet<string>***REMOVED*** */
		this.newBuildDependencies = new LazySet();
		/** @type {Snapshot | undefined***REMOVED*** */
		this.resolveBuildDependenciesSnapshot = undefined;
		/** @type {ResolveResults | undefined***REMOVED*** */
		this.resolveResults = undefined;
		/** @type {Snapshot | undefined***REMOVED*** */
		this.buildSnapshot = undefined;
		/** @type {Promise<Pack> | undefined***REMOVED*** */
		this.packPromise = this._openPack();
		this.storePromise = Promise.resolve();
	***REMOVED***

	/**
	 * @returns {Promise<Pack>***REMOVED*** pack
	 */
	_getPack() {
		if (this.packPromise === undefined) {
			this.packPromise = this.storePromise.then(() => this._openPack());
		***REMOVED***
		return this.packPromise;
	***REMOVED***

	/**
	 * @returns {Promise<Pack>***REMOVED*** the pack
	 */
	_openPack() {
		const { logger, profile, cacheLocation, version ***REMOVED*** = this;
		/** @type {Snapshot***REMOVED*** */
		let buildSnapshot;
		/** @type {BuildDependencies***REMOVED*** */
		let buildDependencies;
		/** @type {BuildDependencies***REMOVED*** */
		let newBuildDependencies;
		/** @type {Snapshot***REMOVED*** */
		let resolveBuildDependenciesSnapshot;
		/** @type {ResolveResults | undefined***REMOVED*** */
		let resolveResults;
		logger.time("restore cache container");
		return this.fileSerializer
			.deserialize(null, {
				filename: `${cacheLocation***REMOVED***/index${this._extension***REMOVED***`,
				extension: `${this._extension***REMOVED***`,
				logger,
				profile,
				retainedBuffer: this.allowCollectingMemory
					? allowCollectingMemory
					: undefined
			***REMOVED***)
			.catch(err => {
				if (err.code !== "ENOENT") {
					logger.warn(
						`Restoring pack failed from ${cacheLocation***REMOVED***${this._extension***REMOVED***: ${err***REMOVED***`
					);
					logger.debug(err.stack);
				***REMOVED*** else {
					logger.debug(
						`No pack exists at ${cacheLocation***REMOVED***${this._extension***REMOVED***: ${err***REMOVED***`
					);
				***REMOVED***
				return undefined;
			***REMOVED***)
			.then(packContainer => {
				logger.timeEnd("restore cache container");
				if (!packContainer) return;
				if (!(packContainer instanceof PackContainer)) {
					logger.warn(
						`Restored pack from ${cacheLocation***REMOVED***${this._extension***REMOVED***, but contained content is unexpected.`,
						packContainer
					);
					return;
				***REMOVED***
				if (packContainer.version !== version) {
					logger.log(
						`Restored pack from ${cacheLocation***REMOVED***${this._extension***REMOVED***, but version doesn't match.`
					);
					return;
				***REMOVED***
				logger.time("check build dependencies");
				return Promise.all([
					new Promise((resolve, reject) => {
						this.fileSystemInfo.checkSnapshotValid(
							packContainer.buildSnapshot,
							(err, valid) => {
								if (err) {
									logger.log(
										`Restored pack from ${cacheLocation***REMOVED***${this._extension***REMOVED***, but checking snapshot of build dependencies errored: ${err***REMOVED***.`
									);
									logger.debug(err.stack);
									return resolve(false);
								***REMOVED***
								if (!valid) {
									logger.log(
										`Restored pack from ${cacheLocation***REMOVED***${this._extension***REMOVED***, but build dependencies have changed.`
									);
									return resolve(false);
								***REMOVED***
								buildSnapshot = packContainer.buildSnapshot;
								return resolve(true);
							***REMOVED***
						);
					***REMOVED***),
					new Promise((resolve, reject) => {
						this.fileSystemInfo.checkSnapshotValid(
							packContainer.resolveBuildDependenciesSnapshot,
							(err, valid) => {
								if (err) {
									logger.log(
										`Restored pack from ${cacheLocation***REMOVED***${this._extension***REMOVED***, but checking snapshot of resolving of build dependencies errored: ${err***REMOVED***.`
									);
									logger.debug(err.stack);
									return resolve(false);
								***REMOVED***
								if (valid) {
									resolveBuildDependenciesSnapshot =
										packContainer.resolveBuildDependenciesSnapshot;
									buildDependencies = packContainer.buildDependencies;
									resolveResults = packContainer.resolveResults;
									return resolve(true);
								***REMOVED***
								logger.log(
									"resolving of build dependencies is invalid, will re-resolve build dependencies"
								);
								this.fileSystemInfo.checkResolveResultsValid(
									packContainer.resolveResults,
									(err, valid) => {
										if (err) {
											logger.log(
												`Restored pack from ${cacheLocation***REMOVED***${this._extension***REMOVED***, but resolving of build dependencies errored: ${err***REMOVED***.`
											);
											logger.debug(err.stack);
											return resolve(false);
										***REMOVED***
										if (valid) {
											newBuildDependencies = packContainer.buildDependencies;
											resolveResults = packContainer.resolveResults;
											return resolve(true);
										***REMOVED***
										logger.log(
											`Restored pack from ${cacheLocation***REMOVED***${this._extension***REMOVED***, but build dependencies resolve to different locations.`
										);
										return resolve(false);
									***REMOVED***
								);
							***REMOVED***
						);
					***REMOVED***)
				])
					.catch(err => {
						logger.timeEnd("check build dependencies");
						throw err;
					***REMOVED***)
					.then(([buildSnapshotValid, resolveValid]) => {
						logger.timeEnd("check build dependencies");
						if (buildSnapshotValid && resolveValid) {
							logger.time("restore cache content metadata");
							const d = /** @type {TODO***REMOVED*** */ (packContainer).data();
							logger.timeEnd("restore cache content metadata");
							return d;
						***REMOVED***
						return undefined;
					***REMOVED***);
			***REMOVED***)
			.then(pack => {
				if (pack) {
					pack.maxAge = this.maxAge;
					this.buildSnapshot = buildSnapshot;
					if (buildDependencies) this.buildDependencies = buildDependencies;
					if (newBuildDependencies)
						this.newBuildDependencies.addAll(newBuildDependencies);
					this.resolveResults = resolveResults;
					this.resolveBuildDependenciesSnapshot =
						resolveBuildDependenciesSnapshot;
					return pack;
				***REMOVED***
				return new Pack(logger, this.maxAge);
			***REMOVED***)
			.catch(err => {
				this.logger.warn(
					`Restoring pack from ${cacheLocation***REMOVED***${this._extension***REMOVED*** failed: ${err***REMOVED***`
				);
				this.logger.debug(err.stack);
				return new Pack(logger, this.maxAge);
			***REMOVED***);
	***REMOVED***

	/**
	 * @param {string***REMOVED*** identifier unique name for the resource
	 * @param {Etag | null***REMOVED*** etag etag of the resource
	 * @param {Data***REMOVED*** data cached content
	 * @returns {Promise<void>***REMOVED*** promise
	 */
	store(identifier, etag, data) {
		if (this.readonly) return Promise.resolve();

		return this._getPack().then(pack => {
			pack.set(identifier, etag === null ? null : etag.toString(), data);
		***REMOVED***);
	***REMOVED***

	/**
	 * @param {string***REMOVED*** identifier unique name for the resource
	 * @param {Etag | null***REMOVED*** etag etag of the resource
	 * @returns {Promise<Data>***REMOVED*** promise to the cached content
	 */
	restore(identifier, etag) {
		return this._getPack()
			.then(pack =>
				pack.get(identifier, etag === null ? null : etag.toString())
			)
			.catch(err => {
				if (err && err.code !== "ENOENT") {
					this.logger.warn(
						`Restoring failed for ${identifier***REMOVED*** from pack: ${err***REMOVED***`
					);
					this.logger.debug(err.stack);
				***REMOVED***
			***REMOVED***);
	***REMOVED***

	/**
	 * @param {LazySet<string> | Iterable<string>***REMOVED*** dependencies dependencies to store
	 */
	storeBuildDependencies(dependencies) {
		if (this.readonly) return;
		this.newBuildDependencies.addAll(dependencies);
	***REMOVED***

	afterAllStored() {
		const packPromise = this.packPromise;
		if (packPromise === undefined) return Promise.resolve();
		const reportProgress = ProgressPlugin.getReporter(this.compiler);
		return (this.storePromise = packPromise
			.then(pack => {
				pack.stopCapturingRequests();
				if (!pack.invalid) return;
				this.packPromise = undefined;
				this.logger.log("Storing pack...");
				let promise;
				const newBuildDependencies = new Set();
				for (const dep of this.newBuildDependencies) {
					if (!this.buildDependencies.has(dep)) {
						newBuildDependencies.add(dep);
					***REMOVED***
				***REMOVED***
				if (newBuildDependencies.size > 0 || !this.buildSnapshot) {
					if (reportProgress) reportProgress(0.5, "resolve build dependencies");
					this.logger.debug(
						`Capturing build dependencies... (${Array.from(
							newBuildDependencies
						).join(", ")***REMOVED***)`
					);
					promise = new Promise(
						/**
						 * @param {(value?: undefined) => void***REMOVED*** resolve resolve
						 * @param {(reason?: Error) => void***REMOVED*** reject reject
						 */
						(resolve, reject) => {
							this.logger.time("resolve build dependencies");
							this.fileSystemInfo.resolveBuildDependencies(
								this.context,
								newBuildDependencies,
								(err, result) => {
									this.logger.timeEnd("resolve build dependencies");
									if (err) return reject(err);

									this.logger.time("snapshot build dependencies");
									const {
										files,
										directories,
										missing,
										resolveResults,
										resolveDependencies
									***REMOVED*** = /** @type {ResolveBuildDependenciesResult***REMOVED*** */ (result);
									if (this.resolveResults) {
										for (const [key, value] of resolveResults) {
											this.resolveResults.set(key, value);
										***REMOVED***
									***REMOVED*** else {
										this.resolveResults = resolveResults;
									***REMOVED***
									if (reportProgress) {
										reportProgress(
											0.6,
											"snapshot build dependencies",
											"resolving"
										);
									***REMOVED***
									this.fileSystemInfo.createSnapshot(
										undefined,
										resolveDependencies.files,
										resolveDependencies.directories,
										resolveDependencies.missing,
										this.snapshot.resolveBuildDependencies,
										(err, snapshot) => {
											if (err) {
												this.logger.timeEnd("snapshot build dependencies");
												return reject(err);
											***REMOVED***
											if (!snapshot) {
												this.logger.timeEnd("snapshot build dependencies");
												return reject(
													new Error("Unable to snapshot resolve dependencies")
												);
											***REMOVED***
											if (this.resolveBuildDependenciesSnapshot) {
												this.resolveBuildDependenciesSnapshot =
													this.fileSystemInfo.mergeSnapshots(
														this.resolveBuildDependenciesSnapshot,
														snapshot
													);
											***REMOVED*** else {
												this.resolveBuildDependenciesSnapshot = snapshot;
											***REMOVED***
											if (reportProgress) {
												reportProgress(
													0.7,
													"snapshot build dependencies",
													"modules"
												);
											***REMOVED***
											this.fileSystemInfo.createSnapshot(
												undefined,
												files,
												directories,
												missing,
												this.snapshot.buildDependencies,
												(err, snapshot) => {
													this.logger.timeEnd("snapshot build dependencies");
													if (err) return reject(err);
													if (!snapshot) {
														return reject(
															new Error("Unable to snapshot build dependencies")
														);
													***REMOVED***
													this.logger.debug("Captured build dependencies");

													if (this.buildSnapshot) {
														this.buildSnapshot =
															this.fileSystemInfo.mergeSnapshots(
																this.buildSnapshot,
																snapshot
															);
													***REMOVED*** else {
														this.buildSnapshot = snapshot;
													***REMOVED***

													resolve();
												***REMOVED***
											);
										***REMOVED***
									);
								***REMOVED***
							);
						***REMOVED***
					);
				***REMOVED*** else {
					promise = Promise.resolve();
				***REMOVED***
				return promise.then(() => {
					if (reportProgress) reportProgress(0.8, "serialize pack");
					this.logger.time("store pack");
					const updatedBuildDependencies = new Set(this.buildDependencies);
					for (const dep of newBuildDependencies) {
						updatedBuildDependencies.add(dep);
					***REMOVED***
					const content = new PackContainer(
						pack,
						this.version,
						/** @type {Snapshot***REMOVED*** */
						(this.buildSnapshot),
						updatedBuildDependencies,
						/** @type {ResolveResults***REMOVED*** */
						(this.resolveResults),
						/** @type {Snapshot***REMOVED*** */
						(this.resolveBuildDependenciesSnapshot)
					);
					return this.fileSerializer
						.serialize(content, {
							filename: `${this.cacheLocation***REMOVED***/index${this._extension***REMOVED***`,
							extension: `${this._extension***REMOVED***`,
							logger: this.logger,
							profile: this.profile
						***REMOVED***)
						.then(() => {
							for (const dep of newBuildDependencies) {
								this.buildDependencies.add(dep);
							***REMOVED***
							this.newBuildDependencies.clear();
							this.logger.timeEnd("store pack");
							const stats = pack.getContentStats();
							this.logger.log(
								"Stored pack (%d items, %d files, %d MiB)",
								pack.itemInfo.size,
								stats.count,
								Math.round(stats.size / 1024 / 1024)
							);
						***REMOVED***)
						.catch(err => {
							this.logger.timeEnd("store pack");
							this.logger.warn(`Caching failed for pack: ${err***REMOVED***`);
							this.logger.debug(err.stack);
						***REMOVED***);
				***REMOVED***);
			***REMOVED***)
			.catch(err => {
				this.logger.warn(`Caching failed for pack: ${err***REMOVED***`);
				this.logger.debug(err.stack);
			***REMOVED***));
	***REMOVED***

	clear() {
		this.fileSystemInfo.clear();
		this.buildDependencies.clear();
		this.newBuildDependencies.clear();
		this.resolveBuildDependenciesSnapshot = undefined;
		this.resolveResults = undefined;
		this.buildSnapshot = undefined;
		this.packPromise = undefined;
	***REMOVED***
***REMOVED***

module.exports = PackFileCacheStrategy;
