// jest-jasmine2/build/expectationResultFactory.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
***REMOVED***);
exports.default = expectationResultFactory;

var _prettyFormat = require('pretty-format');

/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
function messageFormatter({error, message, passed***REMOVED***) {
  if (passed) {
    return 'Passed.';
  ***REMOVED***

  if (message) {
    return message;
  ***REMOVED***

  if (typeof error === 'string') {
    return error;
  ***REMOVED***

  if (
    // duck-type Error, see #2549
    error &&
    typeof error === 'object' &&
    typeof error.message === 'string' &&
    typeof error.name === 'string'
  ) {
    if (error.message === '') {
      return error.name;
    ***REMOVED***

    return `${error.name***REMOVED***: ${error.message***REMOVED***`;
  ***REMOVED***

  return `thrown: ${(0, _prettyFormat.format)(error, {
    maxDepth: 3
  ***REMOVED***)***REMOVED***`;
***REMOVED***

function stackFormatter(options, initError, errorMessage) {
  if (options.passed) {
    return '';
  ***REMOVED***

  if (options.error) {
    if (typeof options.error.stack === 'string') {
      return options.error.stack;
    ***REMOVED***

    if (options.error === errorMessage) {
      return errorMessage;
    ***REMOVED***
  ***REMOVED***

  if (initError) {
    return errorMessage.trimRight() + '\n\n' + initError.stack;
  ***REMOVED***

  return new Error(errorMessage).stack;
***REMOVED***

function expectationResultFactory(options, initError) {
  const message = messageFormatter(options);
  const stack = stackFormatter(options, initError, message);

  if (options.passed) {
    return {
      error: options.error,
      matcherName: options.matcherName,
      message,
      passed: options.passed,
      stack
    ***REMOVED***;
  ***REMOVED***

  return {
    actual: options.actual,
    error: options.error,
    expected: options.expected,
    matcherName: options.matcherName,
    message,
    passed: options.passed,
    stack
  ***REMOVED***;
***REMOVED***
