// workbox-recipes/src/pageCache.ts
/*
  Copyright 2020 Google LLC

  Use of this source code is governed by an MIT-style
  license that can be found in the LICENSE file or at
  https://opensource.org/licenses/MIT.
*/
import {warmStrategyCache***REMOVED*** from './warmStrategyCache';
import {registerRoute***REMOVED*** from 'workbox-routing/registerRoute.js';
import {NetworkFirst***REMOVED*** from 'workbox-strategies/NetworkFirst.js';
import {CacheableResponsePlugin***REMOVED*** from 'workbox-cacheable-response/CacheableResponsePlugin.js';
import {
  RouteMatchCallback,
  RouteMatchCallbackOptions,
  WorkboxPlugin,
***REMOVED*** from 'workbox-core/types.js';

import './_version.js';

export interface PageCacheOptions {
  cacheName?: string;
  matchCallback?: RouteMatchCallback;
  networkTimeoutSeconds?: number;
  plugins?: Array<WorkboxPlugin>;
  warmCache?: Array<string>;
***REMOVED***

/**
 * An implementation of a page caching recipe with a network timeout
 *
 * @memberof workbox-recipes
 *
 * @param {Object***REMOVED*** [options]
 * @param {string***REMOVED*** [options.cacheName] Name for cache. Defaults to pages
 * @param {RouteMatchCallback***REMOVED*** [options.matchCallback] Workbox callback function to call to match to. Defaults to request.mode === 'navigate';
 * @param {number***REMOVED*** [options.networkTimoutSeconds] Maximum amount of time, in seconds, to wait on the network before falling back to cache. Defaults to 3
 * @param {WorkboxPlugin[]***REMOVED*** [options.plugins] Additional plugins to use for this recipe
 * @param {string[]***REMOVED*** [options.warmCache] Paths to call to use to warm this cache
 */
function pageCache(options: PageCacheOptions = {***REMOVED***): void {
  const defaultMatchCallback = ({request***REMOVED***: RouteMatchCallbackOptions) =>
    request.mode === 'navigate';

  const cacheName = options.cacheName || 'pages';
  const matchCallback = options.matchCallback || defaultMatchCallback;
  const networkTimeoutSeconds = options.networkTimeoutSeconds || 3;
  const plugins = options.plugins || [];
  plugins.push(
    new CacheableResponsePlugin({
      statuses: [0, 200],
    ***REMOVED***),
  );

  const strategy = new NetworkFirst({
    networkTimeoutSeconds,
    cacheName,
    plugins,
  ***REMOVED***);

  // Registers the route
  registerRoute(matchCallback, strategy);

  // Warms the cache
  if (options.warmCache) {
    warmStrategyCache({urls: options.warmCache, strategy***REMOVED***);
  ***REMOVED***
***REMOVED***

export {pageCache***REMOVED***;
