// argparse/lib/utils.js
'use strict';

exports.repeat = function (str, num) {
  var result = '';
  for (var i = 0; i < num; i++) { result += str; ***REMOVED***
  return result;
***REMOVED***;

exports.arrayEqual = function (a, b) {
  if (a.length !== b.length) { return false; ***REMOVED***
  for (var i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) { return false; ***REMOVED***
  ***REMOVED***
  return true;
***REMOVED***;

exports.trimChars = function (str, chars) {
  var start = 0;
  var end = str.length - 1;
  while (chars.indexOf(str.charAt(start)) >= 0) { start++; ***REMOVED***
  while (chars.indexOf(str.charAt(end)) >= 0) { end--; ***REMOVED***
  return str.slice(start, end + 1);
***REMOVED***;

exports.capitalize = function (str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
***REMOVED***;

exports.arrayUnion = function () {
  var result = [];
  for (var i = 0, values = {***REMOVED***; i < arguments.length; i++) {
    var arr = arguments[i];
    for (var j = 0; j < arr.length; j++) {
      if (!values[arr[j]]) {
        values[arr[j]] = true;
        result.push(arr[j]);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  return result;
***REMOVED***;

function has(obj, key) {
  return Object.prototype.hasOwnProperty.call(obj, key);
***REMOVED***

exports.has = has;

exports.extend = function (dest, src) {
  for (var i in src) {
    if (has(src, i)) { dest[i] = src[i]; ***REMOVED***
  ***REMOVED***
***REMOVED***;

exports.trimEnd = function (str) {
  return str.replace(/\s+$/g, '');
***REMOVED***;
