// @babel/core/src/config/resolve-targets.ts
type browserType = typeof import("./resolve-targets-browser");
type nodeType = typeof import("./resolve-targets");

// Kind of gross, but essentially asserting that the exports of this module are the same as the
// exports of index-browser, since this file may be replaced at bundle time with index-browser.
({***REMOVED***) as any as browserType as nodeType;

import type { ValidatedOptions ***REMOVED*** from "./validation/options.ts";
import path from "node:path";
import getTargets, {
  type InputTargets,
***REMOVED*** from "@babel/helper-compilation-targets";

import type { Targets ***REMOVED*** from "@babel/helper-compilation-targets";

export function resolveBrowserslistConfigFile(
  browserslistConfigFile: string,
  configFileDir: string,
): string | undefined {
  return path.resolve(configFileDir, browserslistConfigFile);
***REMOVED***

export function resolveTargets(
  options: ValidatedOptions,
  root: string,
): Targets {
  const optTargets = options.targets;
  let targets: InputTargets;

  if (typeof optTargets === "string" || Array.isArray(optTargets)) {
    targets = { browsers: optTargets ***REMOVED***;
  ***REMOVED*** else if (optTargets) {
    if ("esmodules" in optTargets) {
      targets = { ...optTargets, esmodules: "intersect" ***REMOVED***;
    ***REMOVED*** else {
      // https://github.com/microsoft/TypeScript/issues/17002
      targets = optTargets as InputTargets;
    ***REMOVED***
  ***REMOVED***

  const { browserslistConfigFile ***REMOVED*** = options;
  let configFile;
  let ignoreBrowserslistConfig = false;
  if (typeof browserslistConfigFile === "string") {
    configFile = browserslistConfigFile;
  ***REMOVED*** else {
    ignoreBrowserslistConfig = browserslistConfigFile === false;
  ***REMOVED***

  return getTargets(targets, {
    ignoreBrowserslistConfig,
    configFile,
    configPath: root,
    browserslistEnv: options.browserslistEnv,
  ***REMOVED***);
***REMOVED***
