// @babel/preset-env/lib/polyfills/regenerator.cjs
;
const {
  getImportSource,
  getRequireSource
***REMOVED*** = require("./utils.cjs");
function isRegeneratorSource(source) {
  return source === "regenerator-runtime/runtime" || source === "regenerator-runtime/runtime.js";
***REMOVED***
module.exports = function () {
  const visitor = {
    ImportDeclaration(path) {
      if (isRegeneratorSource(getImportSource(path))) {
        this.regeneratorImportExcluded = true;
        path.remove();
      ***REMOVED***
    ***REMOVED***,
    Program(path) {
      path.get("body").forEach(bodyPath => {
        if (isRegeneratorSource(getRequireSource(bodyPath))) {
          this.regeneratorImportExcluded = true;
          bodyPath.remove();
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***;
  return {
    name: "preset-env/remove-regenerator",
    visitor,
    pre() {
      this.regeneratorImportExcluded = false;
    ***REMOVED***,
    post() {
      if (this.opts.debug && this.regeneratorImportExcluded) {
        let filename = this.file.opts.filename;
        if (process.env.BABEL_ENV === "test") {
          filename = filename.replace(/\\/g, "/");
        ***REMOVED***
        console.log(`\n[${filename***REMOVED***] Based on your targets, regenerator-runtime import excluded.`);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;
***REMOVED***;
