// @babel/plugin-transform-regenerator/lib/regenerator/util.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.getTypes = getTypes;
exports.isReference = isReference;
exports.replaceWithOrRemove = replaceWithOrRemove;
exports.runtimeProperty = runtimeProperty;
exports.wrapWithTypes = wrapWithTypes;
let currentTypes = null;
function wrapWithTypes(types, fn) {
  return function (...args) {
    const oldTypes = currentTypes;
    currentTypes = types;
    try {
      return fn.apply(this, args);
    ***REMOVED*** finally {
      currentTypes = oldTypes;
    ***REMOVED***
  ***REMOVED***;
***REMOVED***
function getTypes() {
  return currentTypes;
***REMOVED***
function runtimeProperty(name) {
  const t = getTypes();
  return t.memberExpression(t.identifier("regeneratorRuntime"), t.identifier(name), false);
***REMOVED***
function isReference(path) {
  return path.isReferenced() || path.parentPath.isAssignmentExpression({
    left: path.node
  ***REMOVED***);
***REMOVED***
function replaceWithOrRemove(path, replacement) {
  if (replacement) {
    path.replaceWith(replacement);
  ***REMOVED*** else {
    path.remove();
  ***REMOVED***
***REMOVED***
