// babel-plugin-named-asset-import/index.js
'use strict';

const { extname ***REMOVED*** = require('path');

function namedAssetImportPlugin({ types: t ***REMOVED***) {
  const visited = new WeakSet();

  function generateNewSourcePath(loaderMap, moduleName, sourcePath) {
    const ext = extname(sourcePath).substr(1);
    const extMap = loaderMap[ext];
    return extMap[moduleName]
      ? extMap[moduleName].replace(/\[path\]/, sourcePath)
      : sourcePath;
  ***REMOVED***

  function replaceMatchingSpecifiers(path, loaderMap, callback) {
    const sourcePath = path.node.source.value;
    const ext = extname(sourcePath).substr(1);

    if (visited.has(path.node) || sourcePath.indexOf('!') !== -1) {
      return;
    ***REMOVED***

    if (loaderMap[ext]) {
      path.replaceWithMultiple(
        path.node.specifiers.map(specifier => {
          const newSpecifier = callback(specifier, sourcePath);
          visited.add(newSpecifier);

          return newSpecifier;
        ***REMOVED***)
      );
    ***REMOVED***
  ***REMOVED***

  return {
    visitor: {
      ExportNamedDeclaration(path, { opts: { loaderMap ***REMOVED*** ***REMOVED***) {
        if (!path.node.source) {
          return;
        ***REMOVED***

        replaceMatchingSpecifiers(path, loaderMap, (specifier, sourcePath) => {
          if (t.isExportDefaultSpecifier(specifier)) {
            return t.exportDeclaration(
              [t.exportDefaultSpecifier(t.identifier(specifier.local.name))],
              t.stringLiteral(sourcePath)
            );
          ***REMOVED***

          return t.exportNamedDeclaration(
            null,
            [
              t.exportSpecifier(
                t.identifier(specifier.local.name),
                t.identifier(specifier.exported.name)
              ),
            ],
            t.stringLiteral(
              generateNewSourcePath(loaderMap, specifier.local.name, sourcePath)
            )
          );
        ***REMOVED***);
      ***REMOVED***,
      ImportDeclaration(path, { opts: { loaderMap ***REMOVED*** ***REMOVED***) {
        replaceMatchingSpecifiers(path, loaderMap, (specifier, sourcePath) => {
          if (t.isImportDefaultSpecifier(specifier)) {
            return t.importDeclaration(
              [t.importDefaultSpecifier(t.identifier(specifier.local.name))],
              t.stringLiteral(sourcePath)
            );
          ***REMOVED***

          return t.importDeclaration(
            [
              t.importSpecifier(
                t.identifier(specifier.local.name),
                t.identifier(specifier.imported.name)
              ),
            ],
            t.stringLiteral(
              generateNewSourcePath(
                loaderMap,
                specifier.imported.name,
                sourcePath
              )
            )
          );
        ***REMOVED***);
      ***REMOVED***,
    ***REMOVED***,
  ***REMOVED***;
***REMOVED***

module.exports = namedAssetImportPlugin;
