// jest/build/jest.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
***REMOVED***);
Object.defineProperty(exports, 'SearchSource', {
  enumerable: true,
  get: function () {
    return _core().SearchSource;
  ***REMOVED***
***REMOVED***);
Object.defineProperty(exports, 'TestWatcher', {
  enumerable: true,
  get: function () {
    return _core().TestWatcher;
  ***REMOVED***
***REMOVED***);
Object.defineProperty(exports, 'createTestScheduler', {
  enumerable: true,
  get: function () {
    return _core().createTestScheduler;
  ***REMOVED***
***REMOVED***);
Object.defineProperty(exports, 'getVersion', {
  enumerable: true,
  get: function () {
    return _core().getVersion;
  ***REMOVED***
***REMOVED***);
Object.defineProperty(exports, 'run', {
  enumerable: true,
  get: function () {
    return _jestCli().run;
  ***REMOVED***
***REMOVED***);
Object.defineProperty(exports, 'runCLI', {
  enumerable: true,
  get: function () {
    return _core().runCLI;
  ***REMOVED***
***REMOVED***);

function _core() {
  const data = require('@jest/core');

  _core = function () {
    return data;
  ***REMOVED***;

  return data;
***REMOVED***

function _jestCli() {
  const data = require('jest-cli');

  _jestCli = function () {
    return data;
  ***REMOVED***;

  return data;
***REMOVED***
