// coa/lib/coaobject.js
/* eslint-disable class-methods-use-this */
'use strict';

const Q = require('q');

/**
 * COA Object
 *
 * Base class for all COA-related objects
 *
 * --------|-----|-----|-----
 *         | Cmd | Opt | Arg
 * --------|-----|-----|-----
 *  name   | ✓   | ✓   | ✓
 *  title  | ✓   | ✓   | ✓
 *  comp   | ✓   | ✓   | ✓
 *  reject | ✓   | ✓   | ✓
 *  end    | ✓   | ✓   | ✓
 *  apply  | ✓   | ✓   | ✓
 *
 * @class CoaObject
 */
module.exports = class CoaObject {
    constructor(cmd) {
        this._cmd = cmd;
        this._name = null;
        this._title = null;
        this._comp = null;
    ***REMOVED***

    /**
     * Set a canonical identifier to be used anywhere in the API.
     *
     * @param {String***REMOVED*** name - command, option or argument name
     * @returns {COA.CoaObject***REMOVED*** - this instance (for chainability)
     */
    name(name) {
        this._name = name;
        return this;
    ***REMOVED***

    /**
     * Set a long description to be used anywhere in text messages.
     * @param {String***REMOVED*** title - human readable entity title
     * @returns {COA.CoaObject***REMOVED*** - this instance (for chainability)
     */
    title(title) {
        this._title = title;
        return this;
    ***REMOVED***

    /**
     * Set custom additional completion for current object.
     *
     * @param {Function***REMOVED*** comp - completion generation function,
     *         invoked in the context of object instance.
     *         Accepts parameters:
     *                 - {Object***REMOVED*** opts - completion options
     *         It can return promise or any other value threated as a result.
     * @returns {COA.CoaObject***REMOVED*** - this instance (for chainability)
     */
    comp(comp) {
        this._comp = comp;
        return this;
    ***REMOVED***

    /**
     * Apply function with arguments in a context of object instance.
     *
     * @param {Function***REMOVED*** fn - body
     * @param {Array.<*>***REMOVED*** args... - arguments
     * @returns {COA.CoaObject***REMOVED*** - this instance (for chainability)
     */
    apply(fn) {
        arguments.length > 1?
            fn.apply(this, [].slice.call(arguments, 1))
            : fn.call(this);

        return this;
    ***REMOVED***

    /**
     * Return reject of actions results promise with error code.
     * Use in .act() for return with error.
     * @param {Object***REMOVED*** reason - reject reason
     *         You can customize toString() method and exitCode property
     *         of reason object.
     * @returns {Q.promise***REMOVED*** rejected promise
     */
    reject(reason) {
        return Q.reject(reason);
    ***REMOVED***

    /**
     * Finish chain for current subcommand and return parent command instance.
     * @returns {COA.Cmd***REMOVED*** parent command
     */
    end() {
        return this._cmd;
    ***REMOVED***
***REMOVED***;
