// tapable/lib/AsyncSeriesWaterfallHook.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

const Hook = require("./Hook");
const HookCodeFactory = require("./HookCodeFactory");

class AsyncSeriesWaterfallHookCodeFactory extends HookCodeFactory {
	content({ onError, onResult, onDone ***REMOVED***) {
		return this.callTapsSeries({
			onError: (i, err, next, doneBreak) => onError(err) + doneBreak(true),
			onResult: (i, result, next) => {
				let code = "";
				code += `if(${result***REMOVED*** !== undefined) {\n`;
				code += `${this._args[0]***REMOVED*** = ${result***REMOVED***;\n`;
				code += `***REMOVED***\n`;
				code += next();
				return code;
			***REMOVED***,
			onDone: () => onResult(this._args[0])
		***REMOVED***);
	***REMOVED***
***REMOVED***

const factory = new AsyncSeriesWaterfallHookCodeFactory();

class AsyncSeriesWaterfallHook extends Hook {
	constructor(args) {
		super(args);
		if (args.length < 1)
			throw new Error("Waterfall hooks must have at least one argument");
	***REMOVED***

	compile(options) {
		factory.setup(this, options);
		return factory.create(options);
	***REMOVED***
***REMOVED***

Object.defineProperties(AsyncSeriesWaterfallHook.prototype, {
	_call: { value: undefined, configurable: true, writable: true ***REMOVED***
***REMOVED***);

module.exports = AsyncSeriesWaterfallHook;
