// tapable/lib/AsyncParallelBailHook.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

const Hook = require("./Hook");
const HookCodeFactory = require("./HookCodeFactory");

class AsyncParallelBailHookCodeFactory extends HookCodeFactory {
	content({ onError, onResult, onDone ***REMOVED***) {
		let code = "";
		code += `var _results = new Array(${this.options.taps.length***REMOVED***);\n`;
		code += "var _checkDone = () => {\n";
		code += "for(var i = 0; i < _results.length; i++) {\n";
		code += "var item = _results[i];\n";
		code += "if(item === undefined) return false;\n";
		code += "if(item.result !== undefined) {\n";
		code += onResult("item.result");
		code += "return true;\n";
		code += "***REMOVED***\n";
		code += "if(item.error) {\n";
		code += onError("item.error");
		code += "return true;\n";
		code += "***REMOVED***\n";
		code += "***REMOVED***\n";
		code += "return false;\n";
		code += "***REMOVED***\n";
		code += this.callTapsParallel({
			onError: (i, err, done, doneBreak) => {
				let code = "";
				code += `if(${i***REMOVED*** < _results.length && ((_results.length = ${i +
					1***REMOVED***), (_results[${i***REMOVED***] = { error: ${err***REMOVED*** ***REMOVED***), _checkDone())) {\n`;
				code += doneBreak(true);
				code += "***REMOVED*** else {\n";
				code += done();
				code += "***REMOVED***\n";
				return code;
			***REMOVED***,
			onResult: (i, result, done, doneBreak) => {
				let code = "";
				code += `if(${i***REMOVED*** < _results.length && (${result***REMOVED*** !== undefined && (_results.length = ${i +
					1***REMOVED***), (_results[${i***REMOVED***] = { result: ${result***REMOVED*** ***REMOVED***), _checkDone())) {\n`;
				code += doneBreak(true);
				code += "***REMOVED*** else {\n";
				code += done();
				code += "***REMOVED***\n";
				return code;
			***REMOVED***,
			onTap: (i, run, done, doneBreak) => {
				let code = "";
				if (i > 0) {
					code += `if(${i***REMOVED*** >= _results.length) {\n`;
					code += done();
					code += "***REMOVED*** else {\n";
				***REMOVED***
				code += run();
				if (i > 0) code += "***REMOVED***\n";
				return code;
			***REMOVED***,
			onDone
		***REMOVED***);
		return code;
	***REMOVED***
***REMOVED***

const factory = new AsyncParallelBailHookCodeFactory();

class AsyncParallelBailHook extends Hook {
	compile(options) {
		factory.setup(this, options);
		return factory.create(options);
	***REMOVED***
***REMOVED***

Object.defineProperties(AsyncParallelBailHook.prototype, {
	_call: { value: undefined, configurable: true, writable: true ***REMOVED***
***REMOVED***);

module.exports = AsyncParallelBailHook;
