// tapable/lib/SyncBailHook.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

const Hook = require("./Hook");
const HookCodeFactory = require("./HookCodeFactory");

class SyncBailHookCodeFactory extends HookCodeFactory {
	content({ onError, onResult, resultReturns, onDone, rethrowIfPossible ***REMOVED***) {
		return this.callTapsSeries({
			onError: (i, err) => onError(err),
			onResult: (i, result, next) =>
				`if(${result***REMOVED*** !== undefined) {\n${onResult(
					result
				)***REMOVED***;\n***REMOVED*** else {\n${next()***REMOVED******REMOVED***\n`,
			resultReturns,
			onDone,
			rethrowIfPossible
		***REMOVED***);
	***REMOVED***
***REMOVED***

const factory = new SyncBailHookCodeFactory();

class SyncBailHook extends Hook {
	tapAsync() {
		throw new Error("tapAsync is not supported on a SyncBailHook");
	***REMOVED***

	tapPromise() {
		throw new Error("tapPromise is not supported on a SyncBailHook");
	***REMOVED***

	compile(options) {
		factory.setup(this, options);
		return factory.create(options);
	***REMOVED***
***REMOVED***

module.exports = SyncBailHook;
