// cssnano/src/index.js
'use strict';
const path = require('path');
/** @type {any***REMOVED*** */
const postcss = require('postcss');
const yaml = require('yaml');
const { lilconfigSync ***REMOVED*** = require('lilconfig');

const cssnano = 'cssnano';

/** @typedef {{preset?: any, plugins?: any[], configFile?: string***REMOVED******REMOVED*** Options */
/**
 * @param {string***REMOVED*** moduleId
 * @returns {boolean***REMOVED***
 */
function isResolvable(moduleId) {
  try {
    require.resolve(moduleId);
    return true;
  ***REMOVED*** catch (e) {
    return false;
  ***REMOVED***
***REMOVED***

/**
 * preset can be one of four possibilities:
 * preset = 'default'
 * preset = ['default', {***REMOVED***]
 * preset = function <- to be invoked
 * preset = {plugins: []***REMOVED*** <- already invoked function
 *
 * @param {any***REMOVED*** preset
 * @return {[import('postcss').PluginCreator<any>, boolean | Record<string, any> | undefined][]***REMOVED******REMOVED***
 */
function resolvePreset(preset) {
  let fn, options;

  if (Array.isArray(preset)) {
    fn = preset[0];
    options = preset[1];
  ***REMOVED*** else {
    fn = preset;
    options = {***REMOVED***;
  ***REMOVED***

  // For JS setups where we invoked the preset already
  if (preset.plugins) {
    return preset.plugins;
  ***REMOVED***

  // Provide an alias for the default preset, as it is built-in.
  if (fn === 'default') {
    return require('cssnano-preset-default')(options).plugins;
  ***REMOVED***

  // For non-JS setups; we'll need to invoke the preset ourselves.
  if (typeof fn === 'function') {
    return fn(options).plugins;
  ***REMOVED***

  // Try loading a preset from node_modules
  if (isResolvable(fn)) {
    return require(fn)(options).plugins;
  ***REMOVED***

  const sugar = `cssnano-preset-${fn***REMOVED***`;

  // Try loading a preset from node_modules (sugar)
  if (isResolvable(sugar)) {
    return require(sugar)(options).plugins;
  ***REMOVED***

  // If all else fails, we probably have a typo in the config somewhere
  throw new Error(
    `Cannot load preset "${fn***REMOVED***". Please check your configuration for errors and try again.`
  );
***REMOVED***

/**
 * cssnano will look for configuration firstly as options passed
 * directly to it, and failing this it will use lilconfig to
 * load an external file.

 * @param {Options***REMOVED*** options
 */
function resolveConfig(options) {
  if (options.preset) {
    return resolvePreset(options.preset);
  ***REMOVED***

  /** @type {string | undefined***REMOVED*** */
  let searchPath = process.cwd();
  let configPath = undefined;

  if (options.configFile) {
    searchPath = undefined;
    configPath = path.resolve(process.cwd(), options.configFile);
  ***REMOVED***

  const configExplorer = lilconfigSync(cssnano, {
    searchPlaces: [
      'package.json',
      '.cssnanorc',
      '.cssnanorc.json',
      '.cssnanorc.yaml',
      '.cssnanorc.yml',
      '.cssnanorc.js',
      'cssnano.config.js',
    ],
    loaders: {
      '.yaml': (filepath, content) => yaml.parse(content),
      '.yml': (filepath, content) => yaml.parse(content),
    ***REMOVED***,
  ***REMOVED***);
  const config = configPath
    ? configExplorer.load(configPath)
    : configExplorer.search(searchPath);

  if (config === null) {
    return resolvePreset('default');
  ***REMOVED***

  return resolvePreset(config.config.preset || config.config);
***REMOVED***

/**
 * @type {import('postcss').PluginCreator<Options>***REMOVED***
 * @param {Options=***REMOVED*** options
 * @return {import('postcss').Processor***REMOVED***
 */
function cssnanoPlugin(options = {***REMOVED***) {
  if (Array.isArray(options.plugins)) {
    if (!options.preset || !options.preset.plugins) {
      options.preset = { plugins: [] ***REMOVED***;
    ***REMOVED***

    options.plugins.forEach((plugin) => {
      if (Array.isArray(plugin)) {
        const [pluginDef, opts = {***REMOVED***] = plugin;
        if (typeof pluginDef === 'string' && isResolvable(pluginDef)) {
          options.preset.plugins.push([require(pluginDef), opts]);
        ***REMOVED*** else {
          options.preset.plugins.push([pluginDef, opts]);
        ***REMOVED***
      ***REMOVED*** else if (typeof plugin === 'string' && isResolvable(plugin)) {
        options.preset.plugins.push([require(plugin), {***REMOVED***]);
      ***REMOVED*** else {
        options.preset.plugins.push([plugin, {***REMOVED***]);
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***
  const plugins = [];
  const nanoPlugins = resolveConfig(options);
  for (const nanoPlugin of nanoPlugins) {
    if (Array.isArray(nanoPlugin)) {
      const [processor, opts] = nanoPlugin;
      if (
        typeof opts === 'undefined' ||
        (typeof opts === 'object' && !opts.exclude) ||
        (typeof opts === 'boolean' && opts === true)
      ) {
        plugins.push(processor(opts));
      ***REMOVED***
    ***REMOVED*** else {
      plugins.push(nanoPlugin);
    ***REMOVED***
  ***REMOVED***
  return postcss(plugins);
***REMOVED***

cssnanoPlugin.postcss = true;
module.exports = cssnanoPlugin;
