// core-js-pure/internals/dom-exception-constants.js
'use strict';
module.exports = {
  IndexSizeError: { s: 'INDEX_SIZE_ERR', c: 1, m: 1 ***REMOVED***,
  DOMStringSizeError: { s: 'DOMSTRING_SIZE_ERR', c: 2, m: 0 ***REMOVED***,
  HierarchyRequestError: { s: 'HIERARCHY_REQUEST_ERR', c: 3, m: 1 ***REMOVED***,
  WrongDocumentError: { s: 'WRONG_DOCUMENT_ERR', c: 4, m: 1 ***REMOVED***,
  InvalidCharacterError: { s: 'INVALID_CHARACTER_ERR', c: 5, m: 1 ***REMOVED***,
  NoDataAllowedError: { s: 'NO_DATA_ALLOWED_ERR', c: 6, m: 0 ***REMOVED***,
  NoModificationAllowedError: { s: 'NO_MODIFICATION_ALLOWED_ERR', c: 7, m: 1 ***REMOVED***,
  NotFoundError: { s: 'NOT_FOUND_ERR', c: 8, m: 1 ***REMOVED***,
  NotSupportedError: { s: 'NOT_SUPPORTED_ERR', c: 9, m: 1 ***REMOVED***,
  InUseAttributeError: { s: 'INUSE_ATTRIBUTE_ERR', c: 10, m: 1 ***REMOVED***,
  InvalidStateError: { s: 'INVALID_STATE_ERR', c: 11, m: 1 ***REMOVED***,
  SyntaxError: { s: 'SYNTAX_ERR', c: 12, m: 1 ***REMOVED***,
  InvalidModificationError: { s: 'INVALID_MODIFICATION_ERR', c: 13, m: 1 ***REMOVED***,
  NamespaceError: { s: 'NAMESPACE_ERR', c: 14, m: 1 ***REMOVED***,
  InvalidAccessError: { s: 'INVALID_ACCESS_ERR', c: 15, m: 1 ***REMOVED***,
  ValidationError: { s: 'VALIDATION_ERR', c: 16, m: 0 ***REMOVED***,
  TypeMismatchError: { s: 'TYPE_MISMATCH_ERR', c: 17, m: 1 ***REMOVED***,
  SecurityError: { s: 'SECURITY_ERR', c: 18, m: 1 ***REMOVED***,
  NetworkError: { s: 'NETWORK_ERR', c: 19, m: 1 ***REMOVED***,
  AbortError: { s: 'ABORT_ERR', c: 20, m: 1 ***REMOVED***,
  URLMismatchError: { s: 'URL_MISMATCH_ERR', c: 21, m: 1 ***REMOVED***,
  QuotaExceededError: { s: 'QUOTA_EXCEEDED_ERR', c: 22, m: 1 ***REMOVED***,
  TimeoutError: { s: 'TIMEOUT_ERR', c: 23, m: 1 ***REMOVED***,
  InvalidNodeTypeError: { s: 'INVALID_NODE_TYPE_ERR', c: 24, m: 1 ***REMOVED***,
  DataCloneError: { s: 'DATA_CLONE_ERR', c: 25, m: 1 ***REMOVED***
***REMOVED***;
