// jest-watcher/build/TestWatcher.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
***REMOVED***);
exports.default = void 0;

function _emittery() {
  const data = _interopRequireDefault(require('emittery'));

  _emittery = function () {
    return data;
  ***REMOVED***;

  return data;
***REMOVED***

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {default: obj***REMOVED***;
***REMOVED***

/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
class TestWatcher extends _emittery().default {
  state;
  _isWatchMode;

  constructor({isWatchMode***REMOVED***) {
    super();
    this.state = {
      interrupted: false
    ***REMOVED***;
    this._isWatchMode = isWatchMode;
  ***REMOVED***

  async setState(state) {
    Object.assign(this.state, state);
    await this.emit('change', this.state);
  ***REMOVED***

  isInterrupted() {
    return this.state.interrupted;
  ***REMOVED***

  isWatchMode() {
    return this._isWatchMode;
  ***REMOVED***
***REMOVED***

exports.default = TestWatcher;
