// postcss-normalize-positions/src/index.js
'use strict';
const valueParser = require('postcss-value-parser');

const directionKeywords = new Set(['top', 'right', 'bottom', 'left', 'center']);

const center = '50%';
const horizontal = new Map([
  ['right', '100%'],
  ['left', '0'],
]);
const verticalValue = new Map([
  ['bottom', '100%'],
  ['top', '0'],
]);
const mathFunctions = new Set(['calc', 'min', 'max', 'clamp']);
const variableFunctions = new Set(['var', 'env', 'constant']);
/**
 * @param {valueParser.Node***REMOVED*** node
 * @return {boolean***REMOVED***
 */
function isCommaNode(node) {
  return node.type === 'div' && node.value === ',';
***REMOVED***

/**
 * @param {valueParser.Node***REMOVED*** node
 * @return {boolean***REMOVED***
 */
function isVariableFunctionNode(node) {
  if (node.type !== 'function') {
    return false;
  ***REMOVED***

  return variableFunctions.has(node.value.toLowerCase());
***REMOVED***

/**
 * @param {valueParser.Node***REMOVED*** node
 * @return {boolean***REMOVED***
 */
function isMathFunctionNode(node) {
  if (node.type !== 'function') {
    return false;
  ***REMOVED***
  return mathFunctions.has(node.value.toLowerCase());
***REMOVED***

/**
 * @param {valueParser.Node***REMOVED*** node
 * @return {boolean***REMOVED***
 */
function isNumberNode(node) {
  if (node.type !== 'word') {
    return false;
  ***REMOVED***

  const value = parseFloat(node.value);

  return !isNaN(value);
***REMOVED***

/**
 * @param {valueParser.Node***REMOVED*** node
 * @return {boolean***REMOVED***
 */
function isDimensionNode(node) {
  if (node.type !== 'word') {
    return false;
  ***REMOVED***

  const parsed = valueParser.unit(node.value);

  if (!parsed) {
    return false;
  ***REMOVED***

  return parsed.unit !== '';
***REMOVED***

/**
 * @param {string***REMOVED*** value
 * @return {string***REMOVED***
 */
function transform(value) {
  const parsed = valueParser(value);
  /** @type {({start: number, end: number***REMOVED*** | {start: null, end: null***REMOVED***)[]***REMOVED*** */
  const ranges = [];
  let rangeIndex = 0;
  let shouldContinue = true;

  parsed.nodes.forEach((node, index) => {
    // After comma (`,`) follows next background
    if (isCommaNode(node)) {
      rangeIndex += 1;
      shouldContinue = true;

      return;
    ***REMOVED***

    if (!shouldContinue) {
      return;
    ***REMOVED***

    // After separator (`/`) follows `background-size` values
    // Avoid them
    if (node.type === 'div' && node.value === '/') {
      shouldContinue = false;

      return;
    ***REMOVED***

    if (!ranges[rangeIndex]) {
      ranges[rangeIndex] = {
        start: null,
        end: null,
      ***REMOVED***;
    ***REMOVED***

    // Do not try to be processed `var and `env` function inside background
    if (isVariableFunctionNode(node)) {
      shouldContinue = false;
      ranges[rangeIndex].start = null;
      ranges[rangeIndex].end = null;

      return;
    ***REMOVED***

    const isPositionKeyword =
      (node.type === 'word' &&
        directionKeywords.has(node.value.toLowerCase())) ||
      isDimensionNode(node) ||
      isNumberNode(node) ||
      isMathFunctionNode(node);

    if (ranges[rangeIndex].start === null && isPositionKeyword) {
      ranges[rangeIndex].start = index;
      ranges[rangeIndex].end = index;

      return;
    ***REMOVED***

    if (ranges[rangeIndex].start !== null) {
      if (node.type === 'space') {
        return;
      ***REMOVED*** else if (isPositionKeyword) {
        ranges[rangeIndex].end = index;

        return;
      ***REMOVED***

      return;
    ***REMOVED***
  ***REMOVED***);

  ranges.forEach((range) => {
    if (range.start === null) {
      return;
    ***REMOVED***

    const nodes = parsed.nodes.slice(range.start, range.end + 1);

    if (nodes.length > 3) {
      return;
    ***REMOVED***

    const firstNode = nodes[0].value.toLowerCase();
    const secondNode =
      nodes[2] && nodes[2].value ? nodes[2].value.toLowerCase() : null;

    if (nodes.length === 1 || secondNode === 'center') {
      if (secondNode) {
        nodes[2].value = nodes[1].value = '';
      ***REMOVED***

      const map = new Map([...horizontal, ['center', center]]);

      if (map.has(firstNode)) {
        nodes[0].value = /** @type {string***REMOVED****/ (map.get(firstNode));
      ***REMOVED***

      return;
    ***REMOVED***

    if (secondNode !== null) {
      if (firstNode === 'center' && directionKeywords.has(secondNode)) {
        nodes[0].value = nodes[1].value = '';

        if (horizontal.has(secondNode)) {
          nodes[2].value = /** @type {string***REMOVED*** */ (horizontal.get(secondNode));
        ***REMOVED***
        return;
      ***REMOVED***

      if (horizontal.has(firstNode) && verticalValue.has(secondNode)) {
        nodes[0].value = /** @type {string***REMOVED*** */ (horizontal.get(firstNode));
        nodes[2].value = /** @type {string***REMOVED*** */ (verticalValue.get(secondNode));

        return;
      ***REMOVED*** else if (verticalValue.has(firstNode) && horizontal.has(secondNode)) {
        nodes[0].value = /** @type {string***REMOVED*** */ (horizontal.get(secondNode));
        nodes[2].value = /** @type {string***REMOVED*** */ (verticalValue.get(firstNode));

        return;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***);

  return parsed.toString();
***REMOVED***

/**
 * @type {import('postcss').PluginCreator<void>***REMOVED***
 * @return {import('postcss').Plugin***REMOVED***
 */
function pluginCreator() {
  return {
    postcssPlugin: 'postcss-normalize-positions',

    OnceExit(css) {
      const cache = new Map();

      css.walkDecls(
        /^(background(-position)?|(-\w+-)?perspective-origin)$/i,
        (decl) => {
          const value = decl.value;

          if (!value) {
            return;
          ***REMOVED***

          if (cache.has(value)) {
            decl.value = cache.get(value);

            return;
          ***REMOVED***

          const result = transform(value);

          decl.value = result;
          cache.set(value, result);
        ***REMOVED***
      );
    ***REMOVED***,
  ***REMOVED***;
***REMOVED***

pluginCreator.postcss = true;
module.exports = pluginCreator;
