// jest-validate/build/utils.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
***REMOVED***);
exports.logValidationWarning =
  exports.formatPrettyObject =
  exports.format =
  exports.createDidYouMeanMessage =
  exports.WARNING =
  exports.ValidationError =
  exports.ERROR =
  exports.DEPRECATION =
    void 0;

function _chalk() {
  const data = _interopRequireDefault(require('chalk'));

  _chalk = function () {
    return data;
  ***REMOVED***;

  return data;
***REMOVED***

function _leven() {
  const data = _interopRequireDefault(require('leven'));

  _leven = function () {
    return data;
  ***REMOVED***;

  return data;
***REMOVED***

function _prettyFormat() {
  const data = require('pretty-format');

  _prettyFormat = function () {
    return data;
  ***REMOVED***;

  return data;
***REMOVED***

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {default: obj***REMOVED***;
***REMOVED***

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    ***REMOVED***);
  ***REMOVED*** else {
    obj[key] = value;
  ***REMOVED***
  return obj;
***REMOVED***

const BULLET = _chalk().default.bold('\u25cf');

const DEPRECATION = `${BULLET***REMOVED*** Deprecation Warning`;
exports.DEPRECATION = DEPRECATION;
const ERROR = `${BULLET***REMOVED*** Validation Error`;
exports.ERROR = ERROR;
const WARNING = `${BULLET***REMOVED*** Validation Warning`;
exports.WARNING = WARNING;

const format = value =>
  typeof value === 'function'
    ? value.toString()
    : (0, _prettyFormat().format)(value, {
        min: true
      ***REMOVED***);

exports.format = format;

const formatPrettyObject = value =>
  typeof value === 'function'
    ? value.toString()
    : JSON.stringify(value, null, 2).split('\n').join('\n    ');

exports.formatPrettyObject = formatPrettyObject;

class ValidationError extends Error {
  constructor(name, message, comment) {
    super();

    _defineProperty(this, 'name', void 0);

    _defineProperty(this, 'message', void 0);

    comment = comment ? '\n\n' + comment : '\n';
    this.name = '';
    this.message = _chalk().default.red(
      _chalk().default.bold(name) + ':\n\n' + message + comment
    );
    Error.captureStackTrace(this, () => {***REMOVED***);
  ***REMOVED***
***REMOVED***

exports.ValidationError = ValidationError;

const logValidationWarning = (name, message, comment) => {
  comment = comment ? '\n\n' + comment : '\n';
  console.warn(
    _chalk().default.yellow(
      _chalk().default.bold(name) + ':\n\n' + message + comment
    )
  );
***REMOVED***;

exports.logValidationWarning = logValidationWarning;

const createDidYouMeanMessage = (unrecognized, allowedOptions) => {
  const suggestion = allowedOptions.find(option => {
    const steps = (0, _leven().default)(option, unrecognized);
    return steps < 3;
  ***REMOVED***);
  return suggestion
    ? `Did you mean ${_chalk().default.bold(format(suggestion))***REMOVED***?`
    : '';
***REMOVED***;

exports.createDidYouMeanMessage = createDidYouMeanMessage;
