// merge2/index.js
'use strict'
/*
 * merge2
 * https://github.com/teambition/merge2
 *
 * Copyright (c) 2014-2020 Teambition
 * Licensed under the MIT license.
 */
const Stream = require('stream')
const PassThrough = Stream.PassThrough
const slice = Array.prototype.slice

module.exports = merge2

function merge2 () {
  const streamsQueue = []
  const args = slice.call(arguments)
  let merging = false
  let options = args[args.length - 1]

  if (options && !Array.isArray(options) && options.pipe == null) {
    args.pop()
  ***REMOVED*** else {
    options = {***REMOVED***
  ***REMOVED***

  const doEnd = options.end !== false
  const doPipeError = options.pipeError === true
  if (options.objectMode == null) {
    options.objectMode = true
  ***REMOVED***
  if (options.highWaterMark == null) {
    options.highWaterMark = 64 * 1024
  ***REMOVED***
  const mergedStream = PassThrough(options)

  function addStream () {
    for (let i = 0, len = arguments.length; i < len; i++) {
      streamsQueue.push(pauseStreams(arguments[i], options))
    ***REMOVED***
    mergeStream()
    return this
  ***REMOVED***

  function mergeStream () {
    if (merging) {
      return
    ***REMOVED***
    merging = true

    let streams = streamsQueue.shift()
    if (!streams) {
      process.nextTick(endStream)
      return
    ***REMOVED***
    if (!Array.isArray(streams)) {
      streams = [streams]
    ***REMOVED***

    let pipesCount = streams.length + 1

    function next () {
      if (--pipesCount > 0) {
        return
      ***REMOVED***
      merging = false
      mergeStream()
    ***REMOVED***

    function pipe (stream) {
      function onend () {
        stream.removeListener('merge2UnpipeEnd', onend)
        stream.removeListener('end', onend)
        if (doPipeError) {
          stream.removeListener('error', onerror)
        ***REMOVED***
        next()
      ***REMOVED***
      function onerror (err) {
        mergedStream.emit('error', err)
      ***REMOVED***
      // skip ended stream
      if (stream._readableState.endEmitted) {
        return next()
      ***REMOVED***

      stream.on('merge2UnpipeEnd', onend)
      stream.on('end', onend)

      if (doPipeError) {
        stream.on('error', onerror)
      ***REMOVED***

      stream.pipe(mergedStream, { end: false ***REMOVED***)
      // compatible for old stream
      stream.resume()
    ***REMOVED***

    for (let i = 0; i < streams.length; i++) {
      pipe(streams[i])
    ***REMOVED***

    next()
  ***REMOVED***

  function endStream () {
    merging = false
    // emit 'queueDrain' when all streams merged.
    mergedStream.emit('queueDrain')
    if (doEnd) {
      mergedStream.end()
    ***REMOVED***
  ***REMOVED***

  mergedStream.setMaxListeners(0)
  mergedStream.add = addStream
  mergedStream.on('unpipe', function (stream) {
    stream.emit('merge2UnpipeEnd')
  ***REMOVED***)

  if (args.length) {
    addStream.apply(null, args)
  ***REMOVED***
  return mergedStream
***REMOVED***

// check and pause streams for pipe.
function pauseStreams (streams, options) {
  if (!Array.isArray(streams)) {
    // Backwards-compat with old-style streams
    if (!streams._readableState && streams.pipe) {
      streams = streams.pipe(PassThrough(options))
    ***REMOVED***
    if (!streams._readableState || !streams.pause || !streams.pipe) {
      throw new Error('Only readable stream can be merged.')
    ***REMOVED***
    streams.pause()
  ***REMOVED*** else {
    for (let i = 0, len = streams.length; i < len; i++) {
      streams[i] = pauseStreams(streams[i], options)
    ***REMOVED***
  ***REMOVED***
  return streams
***REMOVED***
