// eslint-plugin-flowtype/dist/rules/booleanStyle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
const schema = [{
  enum: ['bool', 'boolean'],
  type: 'string'
***REMOVED***];

const create = context => {
  const longForm = (context.options[0] || 'boolean') === 'boolean';
  return {
    BooleanTypeAnnotation(node) {
      const diff = node.range[1] - node.range[0];

      if (longForm && diff === 4) {
        context.report({
          fix(fixer) {
            return fixer.replaceText(node, 'boolean');
          ***REMOVED***,

          message: 'Use "boolean", not "bool"',
          node
        ***REMOVED***);
      ***REMOVED***

      if (!longForm && diff !== 4) {
        context.report({
          fix(fixer) {
            return fixer.replaceText(node, 'bool');
          ***REMOVED***,

          message: 'Use "bool", not "boolean"',
          node
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***

  ***REMOVED***;
***REMOVED***;

var _default = {
  create,
  meta: {
    fixable: 'code'
  ***REMOVED***,
  schema
***REMOVED***;
exports.default = _default;
module.exports = exports.default;
