// eslint-plugin-flowtype/dist/rules/interfaceIdMatch.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
const schema = [{
  type: 'string'
***REMOVED***];

const create = context => {
  const pattern = new RegExp(context.options[0] || '^([A-Z][a-z0-9]*)+Type$', 'u');

  const checkInterface = interfaceDeclarationNode => {
    const interfaceIdentifierName = interfaceDeclarationNode.id.name;

    if (!pattern.test(interfaceIdentifierName)) {
      context.report({
        data: {
          name: interfaceIdentifierName,
          pattern: pattern.toString()
        ***REMOVED***,
        message: 'Interface identifier \'{{name***REMOVED******REMOVED***\' does not match pattern \'{{pattern***REMOVED******REMOVED***\'.',
        node: interfaceDeclarationNode
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***;

  return {
    InterfaceDeclaration: checkInterface
  ***REMOVED***;
***REMOVED***;

var _default = {
  create,
  schema
***REMOVED***;
exports.default = _default;
module.exports = exports.default;
