// eslint-plugin-flowtype/dist/rules/noWeakTypes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

var _lodash = _interopRequireDefault(require("lodash"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj ***REMOVED***; ***REMOVED***

const schema = [{
  additionalProperties: false,
  properties: {
    any: {
      type: 'boolean'
    ***REMOVED***,
    Function: {
      type: 'boolean'
    ***REMOVED***,
    Object: {
      type: 'boolean'
    ***REMOVED***
  ***REMOVED***,
  type: 'object'
***REMOVED***];

const reportWeakType = (context, weakType) => {
  return node => {
    context.report({
      data: {
        weakType
      ***REMOVED***,
      message: 'Unexpected use of weak type "{{weakType***REMOVED******REMOVED***"',
      node
    ***REMOVED***);
  ***REMOVED***;
***REMOVED***;

const genericTypeEvaluator = (context, {
  checkFunction,
  checkObject
***REMOVED***) => {
  return node => {
    const name = _lodash.default.get(node, 'id.name');

    if (checkFunction && name === 'Function' || checkObject && name === 'Object') {
      reportWeakType(context, name)(node);
    ***REMOVED***
  ***REMOVED***;
***REMOVED***;

const create = context => {
  const checkAny = _lodash.default.get(context, 'options[0].any', true) === true;
  const checkFunction = _lodash.default.get(context, 'options[0].Function', true) === true;
  const checkObject = _lodash.default.get(context, 'options[0].Object', true) === true;
  const checks = {***REMOVED***;

  if (checkAny) {
    checks.AnyTypeAnnotation = reportWeakType(context, 'any');
  ***REMOVED***

  if (checkFunction || checkObject) {
    checks.GenericTypeAnnotation = genericTypeEvaluator(context, {
      checkFunction,
      checkObject
    ***REMOVED***);
  ***REMOVED***

  return checks;
***REMOVED***;

var _default = {
  create,
  schema
***REMOVED***;
exports.default = _default;
module.exports = exports.default;
