// eslint-plugin-flowtype/dist/rules/quotes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
const schema = [{
  enum: ['double', 'single'],
  type: 'string'
***REMOVED***];
const meta = {
  fixable: 'code'
***REMOVED***;

const create = context => {
  const double = (context.options[0] || 'double') === 'double';
  const sourceCode = context.getSourceCode();
  return {
    StringLiteralTypeAnnotation(node) {
      if (double && sourceCode.text[node.range[0]] !== '"') {
        // double
        context.report({
          fix: fixer => {
            return [fixer.replaceTextRange([node.range[0], node.range[0] + 1], '"'), fixer.replaceTextRange([node.range[1] - 1, node.range[1]], '"')];
          ***REMOVED***,
          message: 'String literals must use double quote.',
          node
        ***REMOVED***);
      ***REMOVED*** else if (!double && sourceCode.text[node.range[0]] !== '\'') {
        // single
        context.report({
          fix: fixer => {
            return [fixer.replaceTextRange([node.range[0], node.range[0] + 1], '\''), fixer.replaceTextRange([node.range[1] - 1, node.range[1]], '\'')];
          ***REMOVED***,
          message: 'String literals must use single quote.',
          node
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***

  ***REMOVED***;
***REMOVED***;

var _default = {
  create,
  meta,
  schema
***REMOVED***;
exports.default = _default;
module.exports = exports.default;
