// eslint-plugin-flowtype/dist/rules/requireParameterType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

var _lodash = _interopRequireDefault(require("lodash"));

var _utilities = require("../utilities");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj ***REMOVED***; ***REMOVED***

const schema = [{
  additionalProperties: false,
  properties: {
    excludeArrowFunctions: {
      enum: [false, true, 'expressionsOnly']
    ***REMOVED***,
    excludeParameterMatch: {
      type: 'string'
    ***REMOVED***
  ***REMOVED***,
  type: 'object'
***REMOVED***];
const create = (0, _utilities.iterateFunctionNodes)(context => {
  const skipArrows = _lodash.default.get(context, 'options[0].excludeArrowFunctions');

  const excludeParameterMatch = new RegExp(_lodash.default.get(context, 'options[0].excludeParameterMatch', 'a^'), 'u');
  return functionNode => {
    // It is save to ignore FunctionTypeAnnotation nodes in this rule.
    if (functionNode.type === 'FunctionTypeAnnotation') {
      return;
    ***REMOVED***

    const isArrow = functionNode.type === 'ArrowFunctionExpression';
    const isArrowFunctionExpression = functionNode.expression;

    const functionAnnotation = isArrow && _lodash.default.get(functionNode, 'parent.id.typeAnnotation');

    if (skipArrows === 'expressionsOnly' && isArrowFunctionExpression || skipArrows === true && isArrow) {
      return;
    ***REMOVED*** // eslint-disable-next-line unicorn/no-array-for-each


    _lodash.default.forEach(functionNode.params, identifierNode => {
      const parameterName = (0, _utilities.getParameterName)(identifierNode, context);

      if (excludeParameterMatch.test(parameterName)) {
        return;
      ***REMOVED***

      let typeAnnotation;
      typeAnnotation = _lodash.default.get(identifierNode, 'typeAnnotation') || _lodash.default.get(identifierNode, 'left.typeAnnotation');

      if (isArrow && functionAnnotation) {
        typeAnnotation = true;
      ***REMOVED***

      if (!typeAnnotation) {
        context.report({
          data: {
            name: (0, _utilities.quoteName)(parameterName)
          ***REMOVED***,
          message: 'Missing {{name***REMOVED******REMOVED***parameter type annotation.',
          node: identifierNode
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***;
***REMOVED***);
var _default = {
  create,
  schema
***REMOVED***;
exports.default = _default;
module.exports = exports.default;
