// eslint-plugin-flowtype/dist/rules/requireReadonlyReactProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

var _lodash = _interopRequireDefault(require("lodash"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj ***REMOVED***; ***REMOVED***

const schema = [{
  additionalProperties: false,
  properties: {
    useImplicitExactTypes: {
      type: 'boolean'
    ***REMOVED***
  ***REMOVED***,
  type: 'object'
***REMOVED***];
const reComponentName = /^(Pure)?Component$/u;
const reReadOnly = /^\$(ReadOnly|FlowFixMe)$/u;

const isReactComponent = node => {
  if (!node.superClass) {
    return false;
  ***REMOVED***

  return (// class Foo extends Component { ***REMOVED***
    // class Foo extends PureComponent { ***REMOVED***
    node.superClass.type === 'Identifier' && reComponentName.test(node.superClass.name) || // class Foo extends React.Component { ***REMOVED***
    // class Foo extends React.PureComponent { ***REMOVED***
    node.superClass.type === 'MemberExpression' && node.superClass.object.name === 'React' && reComponentName.test(node.superClass.property.name)
  );
***REMOVED***; // type Props = {| +foo: string |***REMOVED***


const isReadOnlyObjectType = (node, {
  useImplicitExactTypes
***REMOVED***) => {
  if (!node || node.type !== 'ObjectTypeAnnotation') {
    return false;
  ***REMOVED***

  if (node.properties.length === 0) {
    // we consider `{***REMOVED***` to be ReadOnly since it's exact AND has no props (when `implicitExactTypes=true`)
    // we consider `{||***REMOVED***` to be ReadOnly since it's exact AND has no props (when `implicitExactTypes=false`)
    if (useImplicitExactTypes === true && node.exact === false) {
      return true;
    ***REMOVED***

    if (node.exact === true) {
      return true;
    ***REMOVED***
  ***REMOVED*** // { +foo: ..., +bar: ..., ... ***REMOVED***


  return node.properties.length > 0 && node.properties.every(prop => {
    return prop.variance && prop.variance.kind === 'plus';
  ***REMOVED***);
***REMOVED***; // type Props = {| +foo: string |***REMOVED*** | {| +bar: number |***REMOVED***


const isReadOnlyObjectUnionType = (node, options) => {
  if (!node || node.type !== 'UnionTypeAnnotation') {
    return false;
  ***REMOVED***

  return node.types.every(type => {
    return isReadOnlyObjectType(type, options);
  ***REMOVED***);
***REMOVED***;

const isReadOnlyType = (node, options) => {
  return node.right.id && reReadOnly.test(node.right.id.name) || isReadOnlyObjectType(node.right, options) || isReadOnlyObjectUnionType(node.right, options);
***REMOVED***;

const create = context => {
  const useImplicitExactTypes = _lodash.default.get(context, ['options', 0, 'useImplicitExactTypes'], false);

  const options = {
    useImplicitExactTypes
  ***REMOVED***;
  const readOnlyTypes = [];
  const foundTypes = [];
  const reportedFunctionalComponents = [];

  const isReadOnlyClassProp = node => {
    const id = node.superTypeParameters && node.superTypeParameters.params[0].id;
    return id && !reReadOnly.test(id.name) && !readOnlyTypes.includes(id.name) && foundTypes.includes(id.name);
  ***REMOVED***;

  for (const node of context.getSourceCode().ast.body) {
    let idName;
    let typeNode; // type Props = $ReadOnly<{***REMOVED***>

    if (node.type === 'TypeAlias') {
      idName = node.id.name;
      typeNode = node; // export type Props = $ReadOnly<{***REMOVED***>
    ***REMOVED*** else if (node.type === 'ExportNamedDeclaration' && node.declaration && node.declaration.type === 'TypeAlias') {
      idName = node.declaration.id.name;
      typeNode = node.declaration;
    ***REMOVED***

    if (idName) {
      foundTypes.push(idName);

      if (isReadOnlyType(typeNode, options)) {
        readOnlyTypes.push(idName);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  return {
    // class components
    ClassDeclaration(node) {
      if (isReactComponent(node) && isReadOnlyClassProp(node)) {
        context.report({
          message: node.superTypeParameters.params[0].id.name + ' must be $ReadOnly',
          node
        ***REMOVED***);
      ***REMOVED*** else if (node.superTypeParameters && node.superTypeParameters.params[0].type === 'ObjectTypeAnnotation' && !isReadOnlyObjectType(node.superTypeParameters.params[0], options)) {
        context.report({
          message: node.id.name + ' class props must be $ReadOnly',
          node
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***,

    // functional components
    JSXElement(node) {
      let currentNode = node;
      let identifier;
      let typeAnnotation;

      while (currentNode && currentNode.type !== 'FunctionDeclaration') {
        currentNode = currentNode.parent;
      ***REMOVED*** // functional components can only have 1 param


      if (!currentNode || currentNode.params.length !== 1) {
        return;
      ***REMOVED***

      if (currentNode.params[0].type === 'Identifier' && (typeAnnotation = currentNode.params[0].typeAnnotation)) {
        if ((identifier = typeAnnotation.typeAnnotation.id) && foundTypes.includes(identifier.name) && !readOnlyTypes.includes(identifier.name) && !reReadOnly.test(identifier.name)) {
          if (reportedFunctionalComponents.includes(identifier)) {
            return;
          ***REMOVED***

          context.report({
            message: identifier.name + ' must be $ReadOnly',
            node: identifier
          ***REMOVED***);
          reportedFunctionalComponents.push(identifier);
          return;
        ***REMOVED***

        if (typeAnnotation.typeAnnotation.type === 'ObjectTypeAnnotation' && !isReadOnlyObjectType(typeAnnotation.typeAnnotation, options)) {
          context.report({
            message: currentNode.id.name + ' component props must be $ReadOnly',
            node
          ***REMOVED***);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

  ***REMOVED***;
***REMOVED***;

var _default = {
  create,
  schema
***REMOVED***;
exports.default = _default;
module.exports = exports.default;
