// eslint-plugin-flowtype/dist/rules/requireVariableType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

var _lodash = _interopRequireDefault(require("lodash"));

var _utilities = require("../utilities");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj ***REMOVED***; ***REMOVED***

const schema = [{
  additionalProperties: false,
  properties: {
    excludeVariableMatch: {
      type: 'string'
    ***REMOVED***,
    excludeVariableTypes: {
      additionalProperties: false,
      properties: {
        const: {
          type: 'boolean'
        ***REMOVED***,
        let: {
          type: 'boolean'
        ***REMOVED***,
        var: {
          type: 'boolean'
        ***REMOVED***
      ***REMOVED***,
      type: 'object'
    ***REMOVED***
  ***REMOVED***,
  type: 'object'
***REMOVED***];

const create = context => {
  const checkThisFile = !_lodash.default.get(context, 'settings.flowtype.onlyFilesWithFlowAnnotation') || (0, _utilities.isFlowFile)(context);

  if (!checkThisFile) {
    return () => {***REMOVED***;
  ***REMOVED***

  const excludeVariableMatch = new RegExp(_lodash.default.get(context, 'options[0].excludeVariableMatch', 'a^'), 'u');

  const excludeVariableTypes = _lodash.default.get(context, 'options[0].excludeVariableTypes', {***REMOVED***);

  return {
    VariableDeclaration: variableDeclaration => {
      const variableType = _lodash.default.get(variableDeclaration, 'kind');

      if (_lodash.default.get(excludeVariableTypes, variableType)) {
        return;
      ***REMOVED*** // eslint-disable-next-line unicorn/no-array-for-each


      _lodash.default.forEach(variableDeclaration.declarations, variableDeclarator => {
        const identifierNode = _lodash.default.get(variableDeclarator, 'id');

        const identifierName = _lodash.default.get(identifierNode, 'name');

        if (excludeVariableMatch.test(identifierName)) {
          return;
        ***REMOVED***

        const typeAnnotation = _lodash.default.get(identifierNode, 'typeAnnotation');

        if (!typeAnnotation) {
          context.report({
            data: {
              name: (0, _utilities.quoteName)(identifierName)
            ***REMOVED***,
            message: 'Missing {{name***REMOVED******REMOVED***variable type annotation.',
            node: identifierNode
          ***REMOVED***);
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***;
***REMOVED***;

var _default = {
  create,
  schema
***REMOVED***;
exports.default = _default;
module.exports = exports.default;
