// eslint-plugin-flowtype/dist/rules/requireExactType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
const schema = [{
  enum: ['always', 'never'],
  type: 'string'
***REMOVED***];
const meta = {
  fixable: 'code'
***REMOVED***;

const create = context => {
  const always = (context.options[0] || 'always') === 'always';
  const sourceCode = context.getSourceCode();
  return {
    ObjectTypeAnnotation(node) {
      const {
        exact,
        indexers,
        inexact
      ***REMOVED*** = node;

      if (!['DeclareClass', 'InterfaceDeclaration'].includes(node.parent.type) && always && !exact && !inexact && indexers.length === 0) {
        context.report({
          fix: fixer => {
            return [fixer.replaceText(sourceCode.getFirstToken(node), '{|'), fixer.replaceText(sourceCode.getLastToken(node), '|***REMOVED***')];
          ***REMOVED***,
          message: 'Object type must be exact.',
          node
        ***REMOVED***);
      ***REMOVED***

      if (!always && exact) {
        context.report({
          fix: fixer => {
            return [fixer.replaceText(sourceCode.getFirstToken(node), '{'), fixer.replaceText(sourceCode.getLastToken(node), '***REMOVED***')];
          ***REMOVED***,
          message: 'Object type must not be exact.',
          node
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***

  ***REMOVED***;
***REMOVED***;

var _default = {
  create,
  meta,
  schema
***REMOVED***;
exports.default = _default;
module.exports = exports.default;
