// eslint-plugin-flowtype/dist/rules/semi.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
const schema = [{
  enum: ['always', 'never'],
  type: 'string'
***REMOVED***];

const isSemicolon = token => {
  return token.type === 'Punctuator' && token.value === ';';
***REMOVED***;

const create = context => {
  const never = (context.options[0] || 'always') === 'never';
  const sourceCode = context.getSourceCode();

  const report = (node, missing) => {
    const lastToken = sourceCode.getLastToken(node);
    let fix;
    let message;
    let {
      loc
    ***REMOVED*** = lastToken;

    if (missing) {
      message = 'Missing semicolon.';
      loc = loc.end;

      fix = fixer => {
        return fixer.insertTextAfter(lastToken, ';');
      ***REMOVED***;
    ***REMOVED*** else {
      message = 'Extra semicolon.';
      loc = loc.start;

      fix = fixer => {
        return fixer.remove(lastToken);
      ***REMOVED***;
    ***REMOVED***

    context.report({
      fix,
      loc,
      message,
      node
    ***REMOVED***);
  ***REMOVED***;

  const checkForSemicolon = node => {
    const lastToken = sourceCode.getLastToken(node);
    const isLastTokenSemicolon = isSemicolon(lastToken);

    if (never && isLastTokenSemicolon) {
      report(node, false);
    ***REMOVED***

    if (!never && !isLastTokenSemicolon) {
      report(node, true);
    ***REMOVED***
  ***REMOVED***;

  return {
    OpaqueType: checkForSemicolon,
    TypeAlias: checkForSemicolon,
    TypeAnnotation: node => {
      if (node.parent.type === 'ClassProperty') {
        checkForSemicolon(node.parent);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;
***REMOVED***;

var _default = {
  create,
  meta: {
    fixable: 'code'
  ***REMOVED***,
  schema
***REMOVED***;
exports.default = _default;
module.exports = exports.default;
