// eslint-plugin-flowtype/dist/rules/sortTypeUnionIntersectionMembers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; ***REMOVED***); ***REMOVED*** keys.push.apply(keys, symbols); ***REMOVED*** return keys; ***REMOVED***

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {***REMOVED***; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); ***REMOVED***); ***REMOVED*** else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); ***REMOVED*** else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); ***REMOVED***); ***REMOVED*** ***REMOVED*** return target; ***REMOVED***

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true ***REMOVED***); ***REMOVED*** else { obj[key] = value; ***REMOVED*** return obj; ***REMOVED***

const groups = {
  function: 'function',
  intersection: 'intersection',
  keyword: 'keyword',
  literal: 'literal',
  named: 'named',
  nullish: 'nullish',
  object: 'object',
  tuple: 'tuple',
  union: 'union',
  unknown: 'unknown'
***REMOVED***; // eslint-disable-next-line complexity

const getGroup = node => {
  // eslint-disable-next-line default-case
  switch (node.type) {
    case 'FunctionTypeAnnotation':
      return groups.function;

    case 'IntersectionTypeAnnotation':
      return groups.intersection;

    case 'AnyTypeAnnotation':
    case 'BooleanTypeAnnotation':
    case 'NumberTypeAnnotation':
    case 'StringTypeAnnotation':
    case 'SymbolTypeAnnotation':
    case 'ThisTypeAnnotation':
      return groups.keyword;

    case 'NullLiteralTypeAnnotation':
    case 'NullableTypeAnnotation':
    case 'VoidTypeAnnotation':
      return groups.nullish;

    case 'BooleanLiteralTypeAnnotation':
    case 'NumberLiteralTypeAnnotation':
    case 'StringLiteralTypeAnnotation':
      return groups.literal;

    case 'ArrayTypeAnnotation':
    case 'IndexedAccessType':
    case 'GenericTypeAnnotation':
    case 'OptionalIndexedAccessType':
      return groups.named;

    case 'ObjectTypeAnnotation':
      return groups.object;

    case 'TupleTypeAnnotation':
      return groups.tuple;

    case 'UnionTypeAnnotation':
      return groups.union;
  ***REMOVED***

  return groups.unknown;
***REMOVED***;

const fallbackSort = (a, b) => {
  if (a < b) {
    return -1;
  ***REMOVED***

  if (a > b) {
    return 1;
  ***REMOVED***

  return 0;
***REMOVED***;

const sorters = {
  asc: (collator, a, b) => {
    return collator.compare(a, b) || fallbackSort(a, b);
  ***REMOVED***,
  desc: (collator, a, b) => {
    return collator.compare(b, a) || fallbackSort(b, a);
  ***REMOVED***
***REMOVED***;

const create = context => {
  const sourceCode = context.getSourceCode();
  const {
    checkIntersections = true,
    checkUnions = true,
    groupOrder = [groups.keyword, groups.named, groups.literal, groups.function, groups.object, groups.tuple, groups.intersection, groups.union, groups.nullish],
    order = 'asc'
  ***REMOVED*** = context.options[1] || {***REMOVED***;
  const sort = sorters[order];
  const collator = new Intl.Collator('en', {
    numeric: true,
    sensitivity: 'base'
  ***REMOVED***);

  const checkSorting = node => {
    const sourceOrder = node.types.map(type => {
      var _groupOrder$indexOf;

      const group = (_groupOrder$indexOf = groupOrder === null || groupOrder === void 0 ? void 0 : groupOrder.indexOf(getGroup(type))) !== null && _groupOrder$indexOf !== void 0 ? _groupOrder$indexOf : -1;
      return {
        group: group === -1 ? Number.MAX_SAFE_INTEGER : group,
        node: type,
        text: sourceCode.getText(type)
      ***REMOVED***;
    ***REMOVED***);
    const expectedOrder = [...sourceOrder].sort((a, b) => {
      if (a.group !== b.group) {
        return a.group - b.group;
      ***REMOVED***

      return sort(collator, a.text, b.text);
    ***REMOVED***);
    const hasComments = node.types.some(type => {
      const count = sourceCode.getCommentsBefore(type).length + sourceCode.getCommentsAfter(type).length;
      return count > 0;
    ***REMOVED***);
    let prev = null;

    for (let i = 0; i < expectedOrder.length; i += 1) {
      const type = node.type === 'UnionTypeAnnotation' ? 'union' : 'intersection';
      const current = sourceOrder[i].text;
      const last = prev; // keep track of the last token

      prev = current || last;

      if (!last || !current) {
        continue;
      ***REMOVED***

      if (expectedOrder[i].node !== sourceOrder[i].node) {
        const data = {
          current,
          last,
          order,
          type
        ***REMOVED***;

        const fix = fixer => {
          const sorted = expectedOrder.map(t => {
            return t.text;
          ***REMOVED***).join(node.type === 'UnionTypeAnnotation' ? ' | ' : ' & ');
          return fixer.replaceText(node, sorted);
        ***REMOVED***;

        context.report(_objectSpread({
          data,
          messageId: 'notSorted',
          node
        ***REMOVED***, hasComments ? {
          suggest: [{
            fix,
            messageId: 'suggestFix'
          ***REMOVED***]
        ***REMOVED*** : {
          fix
        ***REMOVED***));
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;

  return {
    IntersectionTypeAnnotation(node) {
      if (checkIntersections === true) {
        checkSorting(node);
      ***REMOVED***
    ***REMOVED***,

    UnionTypeAnnotation(node) {
      if (checkUnions === true) {
        checkSorting(node);
      ***REMOVED***
    ***REMOVED***

  ***REMOVED***;
***REMOVED***;

var _default = {
  create,
  meta: {
    fixable: 'code',
    messages: {
      notSorted: 'Expected {{type***REMOVED******REMOVED*** members to be in {{order***REMOVED******REMOVED***ending order. "{{current***REMOVED******REMOVED***" should be before "{{last***REMOVED******REMOVED***".',
      suggestFix: 'Sort members of type (removes all comments).'
    ***REMOVED***,
    schema: [{
      properties: {
        checkIntersections: {
          type: 'boolean'
        ***REMOVED***,
        checkUnions: {
          type: 'boolean'
        ***REMOVED***,
        groupOrder: {
          items: {
            enum: Object.keys(groups),
            type: 'string'
          ***REMOVED***,
          type: 'array'
        ***REMOVED***,
        order: {
          enum: ['asc', 'desc'],
          type: 'string'
        ***REMOVED***
      ***REMOVED***,
      type: 'object'
    ***REMOVED***]
  ***REMOVED***
***REMOVED***;
exports.default = _default;
module.exports = exports.default;
