// eslint-plugin-flowtype/dist/rules/spaceAfterTypeColon.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

var _lodash = _interopRequireDefault(require("lodash"));

var _typeColonSpacing = _interopRequireDefault(require("./typeColonSpacing"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj ***REMOVED***; ***REMOVED***

const schema = [{
  enum: ['always', 'never'],
  type: 'string'
***REMOVED***, {
  additionalProperties: false,
  properties: {
    allowLineBreak: {
      type: 'boolean'
    ***REMOVED***
  ***REMOVED***,
  type: 'object'
***REMOVED***];

const create = context => {
  return (0, _typeColonSpacing.default)('after', context, {
    allowLineBreak: _lodash.default.get(context, ['options', '1', 'allowLineBreak'], false),
    always: _lodash.default.get(context, ['options', '0'], 'always') === 'always'
  ***REMOVED***);
***REMOVED***;

var _default = {
  create,
  meta: {
    fixable: 'code'
  ***REMOVED***,
  schema
***REMOVED***;
exports.default = _default;
module.exports = exports.default;
