// eslint-plugin-flowtype/dist/rules/spreadExactType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;
const schema = [{
  enum: ['always', 'never'],
  type: 'string'
***REMOVED***];

const create = context => {
  return {
    ObjectTypeAnnotation(node) {
      const {
        properties
      ***REMOVED*** = node;

      for (const property of properties) {
        const {
          type
        ***REMOVED*** = property;

        if (type === 'ObjectTypeSpreadProperty') {
          const {
            argument: {
              type: argumentType,
              id: argumentId
            ***REMOVED***
          ***REMOVED*** = property;

          if (argumentType !== 'GenericTypeAnnotation' || argumentId.name !== '$Exact') {
            context.report({
              message: 'Use $Exact to make type spreading safe.',
              node
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

  ***REMOVED***;
***REMOVED***;

var _default = {
  create,
  schema
***REMOVED***;
exports.default = _default;
module.exports = exports.default;
