// eslint-module-utils/contextCompat.js
'use strict';

exports.__esModule = true;

/** @type {import('./contextCompat').getAncestors***REMOVED*** */
function getAncestors(context, node) {
  const sourceCode = getSourceCode(context);

  if (sourceCode && sourceCode.getAncestors) {
    return sourceCode.getAncestors(node);
  ***REMOVED***

  return context.getAncestors();
***REMOVED***

/** @type {import('./contextCompat').getDeclaredVariables***REMOVED*** */
function getDeclaredVariables(context, node) {
  const sourceCode = getSourceCode(context);

  if (sourceCode && sourceCode.getDeclaredVariables) {
    return sourceCode.getDeclaredVariables(node);
  ***REMOVED***

  return context.getDeclaredVariables(node);
***REMOVED***

/** @type {import('./contextCompat').getFilename***REMOVED*** */
function getFilename(context) {
  if ('filename' in context) {
    return context.filename;
  ***REMOVED***

  return context.getFilename();
***REMOVED***

/** @type {import('./contextCompat').getPhysicalFilename***REMOVED*** */
function getPhysicalFilename(context) {
  if (context.getPhysicalFilename) {
    return context.getPhysicalFilename();
  ***REMOVED***

  return getFilename(context);
***REMOVED***

/** @type {import('./contextCompat').getScope***REMOVED*** */
function getScope(context, node) {
  const sourceCode = getSourceCode(context);

  if (sourceCode && sourceCode.getScope) {
    return sourceCode.getScope(node);
  ***REMOVED***

  return context.getScope();
***REMOVED***

/** @type {import('./contextCompat').getSourceCode***REMOVED*** */
function getSourceCode(context) {
  if ('sourceCode' in context) {
    return context.sourceCode;
  ***REMOVED***

  return context.getSourceCode();
***REMOVED***

module.exports = {
  getAncestors,
  getDeclaredVariables,
  getFilename,
  getPhysicalFilename,
  getScope,
  getSourceCode,
***REMOVED***;
