// eslint-module-utils/moduleVisitor.js
'use strict';

exports.__esModule = true;

/** @typedef {import('estree').Node***REMOVED*** Node */
/** @typedef {{ arguments: import('estree').CallExpression['arguments'], callee: Node ***REMOVED******REMOVED*** Call */
/** @typedef {import('estree').ImportDeclaration | import('estree').ExportNamedDeclaration | import('estree').ExportAllDeclaration***REMOVED*** Declaration */

/**
 * Returns an object of node visitors that will call
 * 'visitor' with every discovered module path.
 *
 * @type {(import('./moduleVisitor').default)***REMOVED***
 */
exports.default = function visitModules(visitor, options) {
  const ignore = options && options.ignore;
  const amd = !!(options && options.amd);
  const commonjs = !!(options && options.commonjs);
  // if esmodule is not explicitly disabled, it is assumed to be enabled
  const esmodule = !!Object.assign({ esmodule: true ***REMOVED***, options).esmodule;

  const ignoreRegExps = ignore == null ? [] : ignore.map((p) => new RegExp(p));

  /** @type {(source: undefined | null | import('estree').Literal, importer: Parameters<typeof visitor>[1]) => void***REMOVED*** */
  function checkSourceValue(source, importer) {
    if (source == null) { return; ***REMOVED*** //?

    // handle ignore
    if (ignoreRegExps.some((re) => re.test(String(source.value)))) { return; ***REMOVED***

    // fire visitor
    visitor(source, importer);
  ***REMOVED***

  // for import-y declarations
  /** @type {(node: Declaration) => void***REMOVED*** */
  function checkSource(node) {
    checkSourceValue(node.source, node);
  ***REMOVED***

  // for esmodule dynamic `import()` calls
  /** @type {(node: import('estree').ImportExpression | import('estree').CallExpression) => void***REMOVED*** */
  function checkImportCall(node) {
    /** @type {import('estree').Expression | import('estree').Literal | import('estree').CallExpression['arguments'][0]***REMOVED*** */
    let modulePath;
    // refs https://github.com/estree/estree/blob/HEAD/es2020.md#importexpression
    if (node.type === 'ImportExpression') {
      modulePath = node.source;
    ***REMOVED*** else if (node.type === 'CallExpression') {
      // @ts-expect-error this structure is from an older version of eslint
      if (node.callee.type !== 'Import') { return; ***REMOVED***
      if (node.arguments.length !== 1) { return; ***REMOVED***

      modulePath = node.arguments[0];
    ***REMOVED*** else {
      throw new TypeError('this should be unreachable');
    ***REMOVED***

    if (modulePath.type !== 'Literal') { return; ***REMOVED***
    if (typeof modulePath.value !== 'string') { return; ***REMOVED***

    checkSourceValue(modulePath, node);
  ***REMOVED***

  // for CommonJS `require` calls
  // adapted from @mctep: https://git.io/v4rAu
  /** @type {(call: Call) => void***REMOVED*** */
  function checkCommon(call) {
    if (call.callee.type !== 'Identifier') { return; ***REMOVED***
    if (call.callee.name !== 'require') { return; ***REMOVED***
    if (call.arguments.length !== 1) { return; ***REMOVED***

    const modulePath = call.arguments[0];
    if (modulePath.type !== 'Literal') { return; ***REMOVED***
    if (typeof modulePath.value !== 'string') { return; ***REMOVED***

    checkSourceValue(modulePath, call);
  ***REMOVED***

  /** @type {(call: Call) => void***REMOVED*** */
  function checkAMD(call) {
    if (call.callee.type !== 'Identifier') { return; ***REMOVED***
    if (call.callee.name !== 'require' && call.callee.name !== 'define') { return; ***REMOVED***
    if (call.arguments.length !== 2) { return; ***REMOVED***

    const modules = call.arguments[0];
    if (modules.type !== 'ArrayExpression') { return; ***REMOVED***

    for (const element of modules.elements) {
      if (!element) { continue; ***REMOVED***
      if (element.type !== 'Literal') { continue; ***REMOVED***
      if (typeof element.value !== 'string') { continue; ***REMOVED***

      if (
        element.value === 'require'
        || element.value === 'exports'
      ) {
        continue; // magic modules: https://github.com/requirejs/requirejs/wiki/Differences-between-the-simplified-CommonJS-wrapper-and-standard-AMD-define#magic-modules
      ***REMOVED***

      checkSourceValue(element, element);
    ***REMOVED***
  ***REMOVED***

  const visitors = {***REMOVED***;
  if (esmodule) {
    Object.assign(visitors, {
      ImportDeclaration: checkSource,
      ExportNamedDeclaration: checkSource,
      ExportAllDeclaration: checkSource,
      CallExpression: checkImportCall,
      ImportExpression: checkImportCall,
    ***REMOVED***);
  ***REMOVED***

  if (commonjs || amd) {
    const currentCallExpression = visitors.CallExpression;
    visitors.CallExpression = /** @type {(call: Call) => void***REMOVED*** */ function (call) {
      if (currentCallExpression) { currentCallExpression(call); ***REMOVED***
      if (commonjs) { checkCommon(call); ***REMOVED***
      if (amd) { checkAMD(call); ***REMOVED***
    ***REMOVED***;
  ***REMOVED***

  return visitors;
***REMOVED***;

/**
 * make an options schema for the module visitor, optionally adding extra fields.
 * @type {import('./moduleVisitor').makeOptionsSchema***REMOVED***
 */
function makeOptionsSchema(additionalProperties) {
  /** @type {import('./moduleVisitor').Schema***REMOVED*** */
  const base =  {
    type: 'object',
    properties: {
      commonjs: { type: 'boolean' ***REMOVED***,
      amd: { type: 'boolean' ***REMOVED***,
      esmodule: { type: 'boolean' ***REMOVED***,
      ignore: {
        type: 'array',
        minItems: 1,
        items: { type: 'string' ***REMOVED***,
        uniqueItems: true,
      ***REMOVED***,
    ***REMOVED***,
    additionalProperties: false,
  ***REMOVED***;

  if (additionalProperties) {
    for (const key in additionalProperties) {
      // @ts-expect-error TS always has trouble with arbitrary object assignment/mutation
      base.properties[key] = additionalProperties[key];
    ***REMOVED***
  ***REMOVED***

  return base;
***REMOVED***
exports.makeOptionsSchema = makeOptionsSchema;

/**
 * json schema object for options parameter. can be used to build rule options schema object.
 */
exports.optionsSchema = makeOptionsSchema();
