// eslint-module-utils/resolve.js
'use strict';

exports.__esModule = true;

const fs = require('fs');
const Module = require('module');
const path = require('path');
const { getPhysicalFilename ***REMOVED*** = require('./contextCompat');

const hashObject = require('./hash').hashObject;
const ModuleCache = require('./ModuleCache').default;
const pkgDir = require('./pkgDir').default;

const CASE_SENSITIVE_FS = !fs.existsSync(path.join(__dirname.toUpperCase(), 'reSOLVE.js'));
exports.CASE_SENSITIVE_FS = CASE_SENSITIVE_FS;

const ERROR_NAME = 'EslintPluginImportResolveError';

const fileExistsCache = new ModuleCache();

// Polyfill Node's `Module.createRequireFromPath` if not present (added in Node v10.12.0)
// Use `Module.createRequire` if available (added in Node v12.2.0)
const createRequire = Module.createRequire
  // @ts-expect-error this only exists in older node
  || Module.createRequireFromPath
  || /** @type {(filename: string) => unknown***REMOVED*** */ function (filename) {
    const mod = new Module(filename, void null);
    mod.filename = filename;
    // @ts-expect-error _nodeModulePaths is undocumented
    mod.paths = Module._nodeModulePaths(path.dirname(filename));

    // @ts-expect-error _compile is undocumented
    mod._compile(`module.exports = require;`, filename);

    return mod.exports;
  ***REMOVED***;

/** @type {(resolver: object) => resolver is import('./resolve').Resolver***REMOVED*** */
function isResolverValid(resolver) {
  if ('interfaceVersion' in resolver && resolver.interfaceVersion === 2) {
    return 'resolve' in resolver && !!resolver.resolve && typeof resolver.resolve === 'function';
  ***REMOVED***
  return 'resolveImport' in resolver && !!resolver.resolveImport && typeof resolver.resolveImport === 'function';
***REMOVED***

/** @type {<T extends string>(target: T, sourceFile?: string | null | undefined) => undefined | ReturnType<typeof require>***REMOVED*** */
function tryRequire(target, sourceFile) {
  let resolved;
  try {
    // Check if the target exists
    if (sourceFile != null) {
      try {
        resolved = createRequire(path.resolve(sourceFile)).resolve(target);
      ***REMOVED*** catch (e) {
        resolved = require.resolve(target);
      ***REMOVED***
    ***REMOVED*** else {
      resolved = require.resolve(target);
    ***REMOVED***
  ***REMOVED*** catch (e) {
    // If the target does not exist then just return undefined
    return undefined;
  ***REMOVED***

  // If the target exists then return the loaded module
  return require(resolved);
***REMOVED***

/** @type {<T extends Map<string, unknown>>(resolvers: string[] | string | { [k: string]: string ***REMOVED***, map: T) => T***REMOVED*** */
function resolverReducer(resolvers, map) {
  if (Array.isArray(resolvers)) {
    resolvers.forEach((r) => resolverReducer(r, map));
    return map;
  ***REMOVED***

  if (typeof resolvers === 'string') {
    map.set(resolvers, null);
    return map;
  ***REMOVED***

  if (typeof resolvers === 'object') {
    for (const key in resolvers) {
      map.set(key, resolvers[key]);
    ***REMOVED***
    return map;
  ***REMOVED***

  const err = new Error('invalid resolver config');
  err.name = ERROR_NAME;
  throw err;
***REMOVED***

/** @type {(sourceFile: string) => string***REMOVED*** */
function getBaseDir(sourceFile) {
  return pkgDir(sourceFile) || process.cwd();
***REMOVED***

/** @type {(name: string, sourceFile: string) => import('./resolve').Resolver***REMOVED*** */
function requireResolver(name, sourceFile) {
  // Try to resolve package with conventional name
  const resolver = tryRequire(`eslint-import-resolver-${name***REMOVED***`, sourceFile)
    || tryRequire(name, sourceFile)
    || tryRequire(path.resolve(getBaseDir(sourceFile), name));

  if (!resolver) {
    const err = new Error(`unable to load resolver "${name***REMOVED***".`);
    err.name = ERROR_NAME;
    throw err;
  ***REMOVED***
  if (!isResolverValid(resolver)) {
    const err = new Error(`${name***REMOVED*** with invalid interface loaded as resolver`);
    err.name = ERROR_NAME;
    throw err;
  ***REMOVED***

  return resolver;
***REMOVED***

// https://stackoverflow.com/a/27382838
/** @type {import('./resolve').fileExistsWithCaseSync***REMOVED*** */
exports.fileExistsWithCaseSync = function fileExistsWithCaseSync(filepath, cacheSettings, strict) {
  // don't care if the FS is case-sensitive
  if (CASE_SENSITIVE_FS) { return true; ***REMOVED***

  // null means it resolved to a builtin
  if (filepath === null) { return true; ***REMOVED***
  if (filepath.toLowerCase() === process.cwd().toLowerCase() && !strict) { return true; ***REMOVED***
  const parsedPath = path.parse(filepath);
  const dir = parsedPath.dir;

  let result = fileExistsCache.get(filepath, cacheSettings);
  if (result != null) { return result; ***REMOVED***

  // base case
  if (dir === '' || parsedPath.root === filepath) {
    result = true;
  ***REMOVED*** else {
    const filenames = fs.readdirSync(dir);
    if (filenames.indexOf(parsedPath.base) === -1) {
      result = false;
    ***REMOVED*** else {
      result = fileExistsWithCaseSync(dir, cacheSettings, strict);
    ***REMOVED***
  ***REMOVED***
  fileExistsCache.set(filepath, result);
  return result;
***REMOVED***;

/** @type {import('./types').ESLintSettings | null***REMOVED*** */
let prevSettings = null;
let memoizedHash = '';
/** @type {(modulePath: string, sourceFile: string, settings: import('./types').ESLintSettings) => import('./resolve').ResolvedResult***REMOVED*** */
function fullResolve(modulePath, sourceFile, settings) {
  // check if this is a bonus core module
  const coreSet = new Set(settings['import/core-modules']);
  if (coreSet.has(modulePath)) { return { found: true, path: null ***REMOVED***; ***REMOVED***

  const sourceDir = path.dirname(sourceFile);

  if (prevSettings !== settings) {
    memoizedHash = hashObject(settings).digest('hex');
    prevSettings = settings;
  ***REMOVED***

  const cacheKey = sourceDir + memoizedHash + modulePath;

  const cacheSettings = ModuleCache.getSettings(settings);

  const cachedPath = fileExistsCache.get(cacheKey, cacheSettings);
  if (cachedPath !== undefined) { return { found: true, path: cachedPath ***REMOVED***; ***REMOVED***

  /** @type {(resolvedPath: string | null) => void***REMOVED*** */
  function cache(resolvedPath) {
    fileExistsCache.set(cacheKey, resolvedPath);
  ***REMOVED***

  /** @type {(resolver: import('./resolve').Resolver, config: unknown) => import('./resolve').ResolvedResult***REMOVED*** */
  function withResolver(resolver, config) {
    if (resolver.interfaceVersion === 2) {
      return resolver.resolve(modulePath, sourceFile, config);
    ***REMOVED***

    try {
      const resolved = resolver.resolveImport(modulePath, sourceFile, config);
      if (resolved === undefined) { return { found: false ***REMOVED***; ***REMOVED***
      return { found: true, path: resolved ***REMOVED***;
    ***REMOVED*** catch (err) {
      return { found: false ***REMOVED***;
    ***REMOVED***
  ***REMOVED***

  const configResolvers = settings['import/resolver']
    || { node: settings['import/resolve'] ***REMOVED***; // backward compatibility

  const resolvers = resolverReducer(configResolvers, new Map());

  for (const pair of resolvers) {
    const name = pair[0];
    const config = pair[1];
    const resolver = requireResolver(name, sourceFile);
    const resolved = withResolver(resolver, config);

    if (!resolved.found) { continue; ***REMOVED***

    // else, counts
    cache(resolved.path);
    return resolved;
  ***REMOVED***

  // failed
  // cache(undefined)
  return { found: false ***REMOVED***;
***REMOVED***

/** @type {import('./resolve').relative***REMOVED*** */
function relative(modulePath, sourceFile, settings) {
  return fullResolve(modulePath, sourceFile, settings).path;
***REMOVED***
exports.relative = relative;

/** @type {Set<import('eslint').Rule.RuleContext>***REMOVED*** */
const erroredContexts = new Set();

/**
 * Given
 * @param p - module path
 * @param context - ESLint context
 * @return - the full module filesystem path; null if package is core; undefined if not found
 * @type {import('./resolve').default***REMOVED***
 */
function resolve(p, context) {
  try {
    return relative(p, getPhysicalFilename(context), context.settings);
  ***REMOVED*** catch (err) {
    if (!erroredContexts.has(context)) {
      // The `err.stack` string starts with `err.name` followed by colon and `err.message`.
      // We're filtering out the default `err.name` because it adds little value to the message.
      // @ts-expect-error this might be an Error
      let errMessage = err.message;
      // @ts-expect-error this might be an Error
      if (err.name !== ERROR_NAME && err.stack) {
        // @ts-expect-error this might be an Error
        errMessage = err.stack.replace(/^Error: /, '');
      ***REMOVED***
      context.report({
        message: `Resolve error: ${errMessage***REMOVED***`,
        loc: { line: 1, column: 0 ***REMOVED***,
      ***REMOVED***);
      erroredContexts.add(context);
    ***REMOVED***
  ***REMOVED***
***REMOVED***
resolve.relative = relative;
exports.default = resolve;
