// table/dist/src/drawBorder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.createTableBorderGetter = exports.drawBorderBottom = exports.drawBorderJoin = exports.drawBorderTop = exports.drawBorder = exports.createSeparatorGetter = exports.drawBorderSegments = void 0;
const drawContent_1 = require("./drawContent");
const drawBorderSegments = (columnWidths, parameters) => {
    const { separator, horizontalBorderIndex, spanningCellManager ***REMOVED*** = parameters;
    return columnWidths.map((columnWidth, columnIndex) => {
        const normalSegment = separator.body.repeat(columnWidth);
        if (horizontalBorderIndex === undefined) {
            return normalSegment;
        ***REMOVED***
        /* istanbul ignore next */
        const range = spanningCellManager === null || spanningCellManager === void 0 ? void 0 : spanningCellManager.getContainingRange({ col: columnIndex,
            row: horizontalBorderIndex ***REMOVED***);
        if (!range) {
            return normalSegment;
        ***REMOVED***
        const { topLeft ***REMOVED*** = range;
        // draw border segments as usual for top border of spanning cell
        if (horizontalBorderIndex === topLeft.row) {
            return normalSegment;
        ***REMOVED***
        // if for first column/row of spanning cell, just skip
        if (columnIndex !== topLeft.col) {
            return '';
        ***REMOVED***
        return range.extractBorderContent(horizontalBorderIndex);
    ***REMOVED***);
***REMOVED***;
exports.drawBorderSegments = drawBorderSegments;
const createSeparatorGetter = (dependencies) => {
    const { separator, spanningCellManager, horizontalBorderIndex, rowCount ***REMOVED*** = dependencies;
    // eslint-disable-next-line complexity
    return (verticalBorderIndex, columnCount) => {
        const inSameRange = spanningCellManager === null || spanningCellManager === void 0 ? void 0 : spanningCellManager.inSameRange;
        if (horizontalBorderIndex !== undefined && inSameRange) {
            const topCell = { col: verticalBorderIndex,
                row: horizontalBorderIndex - 1 ***REMOVED***;
            const leftCell = { col: verticalBorderIndex - 1,
                row: horizontalBorderIndex ***REMOVED***;
            const oppositeCell = { col: verticalBorderIndex - 1,
                row: horizontalBorderIndex - 1 ***REMOVED***;
            const currentCell = { col: verticalBorderIndex,
                row: horizontalBorderIndex ***REMOVED***;
            const pairs = [
                [oppositeCell, topCell],
                [topCell, currentCell],
                [currentCell, leftCell],
                [leftCell, oppositeCell],
            ];
            // left side of horizontal border
            if (verticalBorderIndex === 0) {
                if (inSameRange(currentCell, topCell) && separator.bodyJoinOuter) {
                    return separator.bodyJoinOuter;
                ***REMOVED***
                return separator.left;
            ***REMOVED***
            // right side of horizontal border
            if (verticalBorderIndex === columnCount) {
                if (inSameRange(oppositeCell, leftCell) && separator.bodyJoinOuter) {
                    return separator.bodyJoinOuter;
                ***REMOVED***
                return separator.right;
            ***REMOVED***
            // top horizontal border
            if (horizontalBorderIndex === 0) {
                if (inSameRange(currentCell, leftCell)) {
                    return separator.body;
                ***REMOVED***
                return separator.join;
            ***REMOVED***
            // bottom horizontal border
            if (horizontalBorderIndex === rowCount) {
                if (inSameRange(topCell, oppositeCell)) {
                    return separator.body;
                ***REMOVED***
                return separator.join;
            ***REMOVED***
            const sameRangeCount = pairs.map((pair) => {
                return inSameRange(...pair);
            ***REMOVED***).filter(Boolean).length;
            // four cells are belongs to different spanning cells
            if (sameRangeCount === 0) {
                return separator.join;
            ***REMOVED***
            // belong to one spanning cell
            if (sameRangeCount === 4) {
                return '';
            ***REMOVED***
            // belongs to two spanning cell
            if (sameRangeCount === 2) {
                if (inSameRange(...pairs[1]) && inSameRange(...pairs[3]) && separator.bodyJoinInner) {
                    return separator.bodyJoinInner;
                ***REMOVED***
                return separator.body;
            ***REMOVED***
            /* istanbul ignore next */
            if (sameRangeCount === 1) {
                if (!separator.joinRight || !separator.joinLeft || !separator.joinUp || !separator.joinDown) {
                    throw new Error(`Can not get border separator for position [${horizontalBorderIndex***REMOVED***, ${verticalBorderIndex***REMOVED***]`);
                ***REMOVED***
                if (inSameRange(...pairs[0])) {
                    return separator.joinDown;
                ***REMOVED***
                if (inSameRange(...pairs[1])) {
                    return separator.joinLeft;
                ***REMOVED***
                if (inSameRange(...pairs[2])) {
                    return separator.joinUp;
                ***REMOVED***
                return separator.joinRight;
            ***REMOVED***
            /* istanbul ignore next */
            throw new Error('Invalid case');
        ***REMOVED***
        if (verticalBorderIndex === 0) {
            return separator.left;
        ***REMOVED***
        if (verticalBorderIndex === columnCount) {
            return separator.right;
        ***REMOVED***
        return separator.join;
    ***REMOVED***;
***REMOVED***;
exports.createSeparatorGetter = createSeparatorGetter;
const drawBorder = (columnWidths, parameters) => {
    const borderSegments = (0, exports.drawBorderSegments)(columnWidths, parameters);
    const { drawVerticalLine, horizontalBorderIndex, spanningCellManager ***REMOVED*** = parameters;
    return (0, drawContent_1.drawContent)({
        contents: borderSegments,
        drawSeparator: drawVerticalLine,
        elementType: 'border',
        rowIndex: horizontalBorderIndex,
        separatorGetter: (0, exports.createSeparatorGetter)(parameters),
        spanningCellManager,
    ***REMOVED***) + '\n';
***REMOVED***;
exports.drawBorder = drawBorder;
const drawBorderTop = (columnWidths, parameters) => {
    const { border ***REMOVED*** = parameters;
    const result = (0, exports.drawBorder)(columnWidths, {
        ...parameters,
        separator: {
            body: border.topBody,
            join: border.topJoin,
            left: border.topLeft,
            right: border.topRight,
        ***REMOVED***,
    ***REMOVED***);
    if (result === '\n') {
        return '';
    ***REMOVED***
    return result;
***REMOVED***;
exports.drawBorderTop = drawBorderTop;
const drawBorderJoin = (columnWidths, parameters) => {
    const { border ***REMOVED*** = parameters;
    return (0, exports.drawBorder)(columnWidths, {
        ...parameters,
        separator: {
            body: border.joinBody,
            bodyJoinInner: border.bodyJoin,
            bodyJoinOuter: border.bodyLeft,
            join: border.joinJoin,
            joinDown: border.joinMiddleDown,
            joinLeft: border.joinMiddleLeft,
            joinRight: border.joinMiddleRight,
            joinUp: border.joinMiddleUp,
            left: border.joinLeft,
            right: border.joinRight,
        ***REMOVED***,
    ***REMOVED***);
***REMOVED***;
exports.drawBorderJoin = drawBorderJoin;
const drawBorderBottom = (columnWidths, parameters) => {
    const { border ***REMOVED*** = parameters;
    return (0, exports.drawBorder)(columnWidths, {
        ...parameters,
        separator: {
            body: border.bottomBody,
            join: border.bottomJoin,
            left: border.bottomLeft,
            right: border.bottomRight,
        ***REMOVED***,
    ***REMOVED***);
***REMOVED***;
exports.drawBorderBottom = drawBorderBottom;
const createTableBorderGetter = (columnWidths, parameters) => {
    return (index, size) => {
        const drawBorderParameters = { ...parameters,
            horizontalBorderIndex: index ***REMOVED***;
        if (index === 0) {
            return (0, exports.drawBorderTop)(columnWidths, drawBorderParameters);
        ***REMOVED***
        else if (index === size) {
            return (0, exports.drawBorderBottom)(columnWidths, drawBorderParameters);
        ***REMOVED***
        return (0, exports.drawBorderJoin)(columnWidths, drawBorderParameters);
    ***REMOVED***;
***REMOVED***;
exports.createTableBorderGetter = createTableBorderGetter;
