// table/dist/src/injectHeaderConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.injectHeaderConfig = void 0;
const injectHeaderConfig = (rows, config) => {
    var _a;
    let spanningCellConfig = (_a = config.spanningCells) !== null && _a !== void 0 ? _a : [];
    const headerConfig = config.header;
    const adjustedRows = [...rows];
    if (headerConfig) {
        spanningCellConfig = spanningCellConfig.map(({ row, ...rest ***REMOVED***) => {
            return { ...rest,
                row: row + 1 ***REMOVED***;
        ***REMOVED***);
        const { content, ...headerStyles ***REMOVED*** = headerConfig;
        spanningCellConfig.unshift({ alignment: 'center',
            col: 0,
            colSpan: rows[0].length,
            paddingLeft: 1,
            paddingRight: 1,
            row: 0,
            wrapWord: false,
            ...headerStyles ***REMOVED***);
        adjustedRows.unshift([content, ...Array.from({ length: rows[0].length - 1 ***REMOVED***).fill('')]);
    ***REMOVED***
    return [adjustedRows,
        spanningCellConfig];
***REMOVED***;
exports.injectHeaderConfig = injectHeaderConfig;
