// @eslint/eslintrc/conf/config-schema.js
/**
 * @fileoverview Defines a schema for configs.
 * @author Sylvan Mably
 */

const baseConfigProperties = {
    $schema: { type: "string" ***REMOVED***,
    env: { type: "object" ***REMOVED***,
    extends: { $ref: "#/definitions/stringOrStrings" ***REMOVED***,
    globals: { type: "object" ***REMOVED***,
    overrides: {
        type: "array",
        items: { $ref: "#/definitions/overrideConfig" ***REMOVED***,
        additionalItems: false
    ***REMOVED***,
    parser: { type: ["string", "null"] ***REMOVED***,
    parserOptions: { type: "object" ***REMOVED***,
    plugins: { type: "array" ***REMOVED***,
    processor: { type: "string" ***REMOVED***,
    rules: { type: "object" ***REMOVED***,
    settings: { type: "object" ***REMOVED***,
    noInlineConfig: { type: "boolean" ***REMOVED***,
    reportUnusedDisableDirectives: { type: "boolean" ***REMOVED***,

    ecmaFeatures: { type: "object" ***REMOVED*** // deprecated; logs a warning when used
***REMOVED***;

const configSchema = {
    definitions: {
        stringOrStrings: {
            oneOf: [
                { type: "string" ***REMOVED***,
                {
                    type: "array",
                    items: { type: "string" ***REMOVED***,
                    additionalItems: false
                ***REMOVED***
            ]
        ***REMOVED***,
        stringOrStringsRequired: {
            oneOf: [
                { type: "string" ***REMOVED***,
                {
                    type: "array",
                    items: { type: "string" ***REMOVED***,
                    additionalItems: false,
                    minItems: 1
                ***REMOVED***
            ]
        ***REMOVED***,

        // Config at top-level.
        objectConfig: {
            type: "object",
            properties: {
                root: { type: "boolean" ***REMOVED***,
                ignorePatterns: { $ref: "#/definitions/stringOrStrings" ***REMOVED***,
                ...baseConfigProperties
            ***REMOVED***,
            additionalProperties: false
        ***REMOVED***,

        // Config in `overrides`.
        overrideConfig: {
            type: "object",
            properties: {
                excludedFiles: { $ref: "#/definitions/stringOrStrings" ***REMOVED***,
                files: { $ref: "#/definitions/stringOrStringsRequired" ***REMOVED***,
                ...baseConfigProperties
            ***REMOVED***,
            required: ["files"],
            additionalProperties: false
        ***REMOVED***
    ***REMOVED***,

    $ref: "#/definitions/objectConfig"
***REMOVED***;

export default configSchema;
