// postcss-normalize-url/src/index.js
'use strict';
const path = require('path');
const valueParser = require('postcss-value-parser');
const normalize = require('normalize-url');

const multiline = /\\[\r\n]/;
// eslint-disable-next-line no-useless-escape
const escapeChars = /([\s\(\)"'])/g;

// Scheme: https://tools.ietf.org/html/rfc3986#section-3.1
// Absolute URL: https://tools.ietf.org/html/rfc3986#section-4.3
const ABSOLUTE_URL_REGEX = /^[a-zA-Z][a-zA-Z\d+\-.]*?:/;
// Windows paths like `c:\`
const WINDOWS_PATH_REGEX = /^[a-zA-Z]:\\/;

/**
 * Originally in sindresorhus/is-absolute-url
 *
 * @param {string***REMOVED*** url
 */
function isAbsolute(url) {
  if (WINDOWS_PATH_REGEX.test(url)) {
    return false;
  ***REMOVED***
  return ABSOLUTE_URL_REGEX.test(url);
***REMOVED***

/**
 * @param {string***REMOVED*** url
 * @param {normalize.Options***REMOVED*** options
 * @return {string***REMOVED***
 */
function convert(url, options) {
  if (isAbsolute(url) || url.startsWith('//')) {
    let normalizedURL;

    try {
      normalizedURL = normalize(url, options);
    ***REMOVED*** catch (e) {
      normalizedURL = url;
    ***REMOVED***

    return normalizedURL;
  ***REMOVED***

  // `path.normalize` always returns backslashes on Windows, need replace in `/`
  return path.normalize(url).replace(new RegExp('\\' + path.sep, 'g'), '/');
***REMOVED***

/**
 * @param {import('postcss').AtRule***REMOVED*** rule
 * @return {void***REMOVED***
 */
function transformNamespace(rule) {
  rule.params = valueParser(rule.params)
    .walk((node) => {
      if (
        node.type === 'function' &&
        node.value.toLowerCase() === 'url' &&
        node.nodes.length
      ) {
        /** @type {valueParser.Node***REMOVED*** */ (node).type = 'string';
        /** @type {any***REMOVED*** */ (node).quote =
          node.nodes[0].type === 'string' ? node.nodes[0].quote : '"';
        node.value = node.nodes[0].value;
      ***REMOVED***
      if (node.type === 'string') {
        node.value = node.value.trim();
      ***REMOVED***
      return false;
    ***REMOVED***)
    .toString();
***REMOVED***

/**
 * @param {import('postcss').Declaration***REMOVED*** decl
 * @param {normalize.Options***REMOVED*** opts
 * @return {void***REMOVED***
 */
function transformDecl(decl, opts) {
  decl.value = valueParser(decl.value)
    .walk((node) => {
      if (node.type !== 'function' || node.value.toLowerCase() !== 'url') {
        return false;
      ***REMOVED***

      node.before = node.after = '';

      if (!node.nodes.length) {
        return false;
      ***REMOVED***
      let url = node.nodes[0];
      let escaped;

      url.value = url.value.trim().replace(multiline, '');

      // Skip empty URLs
      // Empty URL function equals request to current stylesheet where it is declared
      if (url.value.length === 0) {
        /** @type {any***REMOVED*** */ (url).quote = '';

        return false;
      ***REMOVED***

      if (/^data:(.*)?,/i.test(url.value)) {
        return false;
      ***REMOVED***

      if (!/^.+-extension:\//i.test(url.value)) {
        url.value = convert(url.value, opts);
      ***REMOVED***

      if (escapeChars.test(url.value) && url.type === 'string') {
        escaped = url.value.replace(escapeChars, '\\$1');

        if (escaped.length < url.value.length + 2) {
          url.value = escaped;
          /** @type {valueParser.Node***REMOVED*** */ (url).type = 'word';
        ***REMOVED***
      ***REMOVED*** else {
        url.type = 'word';
      ***REMOVED***

      return false;
    ***REMOVED***)
    .toString();
***REMOVED***

/** @typedef {normalize.Options***REMOVED*** Options */
/**
 * @type {import('postcss').PluginCreator<Options>***REMOVED***
 * @param {Options***REMOVED*** opts
 * @return {import('postcss').Plugin***REMOVED***
 */
function pluginCreator(opts) {
  opts = Object.assign(
    {***REMOVED***,
    {
      normalizeProtocol: false,
      sortQueryParameters: false,
      stripHash: false,
      stripWWW: false,
      stripTextFragment: false,
    ***REMOVED***,
    opts
  );

  return {
    postcssPlugin: 'postcss-normalize-url',

    OnceExit(css) {
      css.walk((node) => {
        if (node.type === 'decl') {
          return transformDecl(node, opts);
        ***REMOVED*** else if (
          node.type === 'atrule' &&
          node.name.toLowerCase() === 'namespace'
        ) {
          return transformNamespace(node);
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***,
  ***REMOVED***;
***REMOVED***

pluginCreator.postcss = true;
module.exports = pluginCreator;
