// @babel/traverse/lib/path/inference/inferer-reference.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = _default;
var _t = require("@babel/types");
var _util = require("./util.js");
const {
  BOOLEAN_NUMBER_BINARY_OPERATORS,
  createTypeAnnotationBasedOnTypeof,
  numberTypeAnnotation,
  voidTypeAnnotation
***REMOVED*** = _t;
function _default(node) {
  if (!this.isReferenced()) return;
  const binding = this.scope.getBinding(node.name);
  if (binding) {
    if (binding.identifier.typeAnnotation) {
      return binding.identifier.typeAnnotation;
    ***REMOVED*** else {
      return getTypeAnnotationBindingConstantViolations(binding, this, node.name);
    ***REMOVED***
  ***REMOVED***
  if (node.name === "undefined") {
    return voidTypeAnnotation();
  ***REMOVED*** else if (node.name === "NaN" || node.name === "Infinity") {
    return numberTypeAnnotation();
  ***REMOVED*** else if (node.name === "arguments") {***REMOVED***
***REMOVED***
function getTypeAnnotationBindingConstantViolations(binding, path, name) {
  const types = [];
  const functionConstantViolations = [];
  let constantViolations = getConstantViolationsBefore(binding, path, functionConstantViolations);
  const testType = getConditionalAnnotation(binding, path, name);
  if (testType) {
    const testConstantViolations = getConstantViolationsBefore(binding, testType.ifStatement);
    constantViolations = constantViolations.filter(path => !testConstantViolations.includes(path));
    types.push(testType.typeAnnotation);
  ***REMOVED***
  if (constantViolations.length) {
    constantViolations.push(...functionConstantViolations);
    for (const violation of constantViolations) {
      types.push(violation.getTypeAnnotation());
    ***REMOVED***
  ***REMOVED***
  if (!types.length) {
    return;
  ***REMOVED***
  return (0, _util.createUnionType)(types);
***REMOVED***
function getConstantViolationsBefore(binding, path, functions) {
  const violations = binding.constantViolations.slice();
  violations.unshift(binding.path);
  return violations.filter(violation => {
    violation = violation.resolve();
    const status = violation._guessExecutionStatusRelativeTo(path);
    if (functions && status === "unknown") functions.push(violation);
    return status === "before";
  ***REMOVED***);
***REMOVED***
function inferAnnotationFromBinaryExpression(name, path) {
  const operator = path.node.operator;
  const right = path.get("right").resolve();
  const left = path.get("left").resolve();
  let target;
  if (left.isIdentifier({
    name
  ***REMOVED***)) {
    target = right;
  ***REMOVED*** else if (right.isIdentifier({
    name
  ***REMOVED***)) {
    target = left;
  ***REMOVED***
  if (target) {
    if (operator === "===") {
      return target.getTypeAnnotation();
    ***REMOVED***
    if (BOOLEAN_NUMBER_BINARY_OPERATORS.includes(operator)) {
      return numberTypeAnnotation();
    ***REMOVED***
    return;
  ***REMOVED***
  if (operator !== "===" && operator !== "==") return;
  let typeofPath;
  let typePath;
  if (left.isUnaryExpression({
    operator: "typeof"
  ***REMOVED***)) {
    typeofPath = left;
    typePath = right;
  ***REMOVED*** else if (right.isUnaryExpression({
    operator: "typeof"
  ***REMOVED***)) {
    typeofPath = right;
    typePath = left;
  ***REMOVED***
  if (!typeofPath) return;
  if (!typeofPath.get("argument").isIdentifier({
    name
  ***REMOVED***)) return;
  typePath = typePath.resolve();
  if (!typePath.isLiteral()) return;
  const typeValue = typePath.node.value;
  if (typeof typeValue !== "string") return;
  return createTypeAnnotationBasedOnTypeof(typeValue);
***REMOVED***
function getParentConditionalPath(binding, path, name) {
  let parentPath;
  while (parentPath = path.parentPath) {
    if (parentPath.isIfStatement() || parentPath.isConditionalExpression()) {
      if (path.key === "test") {
        return;
      ***REMOVED***
      return parentPath;
    ***REMOVED***
    if (parentPath.isFunction()) {
      if (parentPath.parentPath.scope.getBinding(name) !== binding) return;
    ***REMOVED***
    path = parentPath;
  ***REMOVED***
***REMOVED***
function getConditionalAnnotation(binding, path, name) {
  const ifStatement = getParentConditionalPath(binding, path, name);
  if (!ifStatement) return;
  const test = ifStatement.get("test");
  const paths = [test];
  const types = [];
  for (let i = 0; i < paths.length; i++) {
    const path = paths[i];
    if (path.isLogicalExpression()) {
      if (path.node.operator === "&&") {
        paths.push(path.get("left"));
        paths.push(path.get("right"));
      ***REMOVED***
    ***REMOVED*** else if (path.isBinaryExpression()) {
      const type = inferAnnotationFromBinaryExpression(name, path);
      if (type) types.push(type);
    ***REMOVED***
  ***REMOVED***
  if (types.length) {
    return {
      typeAnnotation: (0, _util.createUnionType)(types),
      ifStatement
    ***REMOVED***;
  ***REMOVED***
  return getConditionalAnnotation(binding, ifStatement, name);
***REMOVED***
