// @babel/traverse/lib/path/inference/inferers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.ArrayExpression = ArrayExpression;
exports.AssignmentExpression = AssignmentExpression;
exports.BinaryExpression = BinaryExpression;
exports.BooleanLiteral = BooleanLiteral;
exports.CallExpression = CallExpression;
exports.ConditionalExpression = ConditionalExpression;
exports.ClassDeclaration = exports.ClassExpression = exports.FunctionDeclaration = exports.ArrowFunctionExpression = exports.FunctionExpression = Func;
Object.defineProperty(exports, "Identifier", {
  enumerable: true,
  get: function () {
    return _infererReference.default;
  ***REMOVED***
***REMOVED***);
exports.LogicalExpression = LogicalExpression;
exports.NewExpression = NewExpression;
exports.NullLiteral = NullLiteral;
exports.NumericLiteral = NumericLiteral;
exports.ObjectExpression = ObjectExpression;
exports.ParenthesizedExpression = ParenthesizedExpression;
exports.RegExpLiteral = RegExpLiteral;
exports.RestElement = RestElement;
exports.SequenceExpression = SequenceExpression;
exports.StringLiteral = StringLiteral;
exports.TSAsExpression = TSAsExpression;
exports.TSNonNullExpression = TSNonNullExpression;
exports.TaggedTemplateExpression = TaggedTemplateExpression;
exports.TemplateLiteral = TemplateLiteral;
exports.TypeCastExpression = TypeCastExpression;
exports.UnaryExpression = UnaryExpression;
exports.UpdateExpression = UpdateExpression;
exports.VariableDeclarator = VariableDeclarator;
var _t = require("@babel/types");
var _infererReference = require("./inferer-reference.js");
var _util = require("./util.js");
const {
  BOOLEAN_BINARY_OPERATORS,
  BOOLEAN_UNARY_OPERATORS,
  NUMBER_BINARY_OPERATORS,
  NUMBER_UNARY_OPERATORS,
  STRING_UNARY_OPERATORS,
  anyTypeAnnotation,
  arrayTypeAnnotation,
  booleanTypeAnnotation,
  buildMatchMemberExpression,
  genericTypeAnnotation,
  identifier,
  nullLiteralTypeAnnotation,
  numberTypeAnnotation,
  stringTypeAnnotation,
  tupleTypeAnnotation,
  unionTypeAnnotation,
  voidTypeAnnotation,
  isIdentifier
***REMOVED*** = _t;
function VariableDeclarator() {
  if (!this.get("id").isIdentifier()) return;
  return this.get("init").getTypeAnnotation();
***REMOVED***
function TypeCastExpression(node) {
  return node.typeAnnotation;
***REMOVED***
TypeCastExpression.validParent = true;
function TSAsExpression(node) {
  return node.typeAnnotation;
***REMOVED***
TSAsExpression.validParent = true;
function TSNonNullExpression() {
  return this.get("expression").getTypeAnnotation();
***REMOVED***
function NewExpression(node) {
  if (node.callee.type === "Identifier") {
    return genericTypeAnnotation(node.callee);
  ***REMOVED***
***REMOVED***
function TemplateLiteral() {
  return stringTypeAnnotation();
***REMOVED***
function UnaryExpression(node) {
  const operator = node.operator;
  if (operator === "void") {
    return voidTypeAnnotation();
  ***REMOVED*** else if (NUMBER_UNARY_OPERATORS.includes(operator)) {
    return numberTypeAnnotation();
  ***REMOVED*** else if (STRING_UNARY_OPERATORS.includes(operator)) {
    return stringTypeAnnotation();
  ***REMOVED*** else if (BOOLEAN_UNARY_OPERATORS.includes(operator)) {
    return booleanTypeAnnotation();
  ***REMOVED***
***REMOVED***
function BinaryExpression(node) {
  const operator = node.operator;
  if (NUMBER_BINARY_OPERATORS.includes(operator)) {
    return numberTypeAnnotation();
  ***REMOVED*** else if (BOOLEAN_BINARY_OPERATORS.includes(operator)) {
    return booleanTypeAnnotation();
  ***REMOVED*** else if (operator === "+") {
    const right = this.get("right");
    const left = this.get("left");
    if (left.isBaseType("number") && right.isBaseType("number")) {
      return numberTypeAnnotation();
    ***REMOVED*** else if (left.isBaseType("string") || right.isBaseType("string")) {
      return stringTypeAnnotation();
    ***REMOVED***
    return unionTypeAnnotation([stringTypeAnnotation(), numberTypeAnnotation()]);
  ***REMOVED***
***REMOVED***
function LogicalExpression() {
  const argumentTypes = [this.get("left").getTypeAnnotation(), this.get("right").getTypeAnnotation()];
  return (0, _util.createUnionType)(argumentTypes);
***REMOVED***
function ConditionalExpression() {
  const argumentTypes = [this.get("consequent").getTypeAnnotation(), this.get("alternate").getTypeAnnotation()];
  return (0, _util.createUnionType)(argumentTypes);
***REMOVED***
function SequenceExpression() {
  return this.get("expressions").pop().getTypeAnnotation();
***REMOVED***
function ParenthesizedExpression() {
  return this.get("expression").getTypeAnnotation();
***REMOVED***
function AssignmentExpression() {
  return this.get("right").getTypeAnnotation();
***REMOVED***
function UpdateExpression(node) {
  const operator = node.operator;
  if (operator === "++" || operator === "--") {
    return numberTypeAnnotation();
  ***REMOVED***
***REMOVED***
function StringLiteral() {
  return stringTypeAnnotation();
***REMOVED***
function NumericLiteral() {
  return numberTypeAnnotation();
***REMOVED***
function BooleanLiteral() {
  return booleanTypeAnnotation();
***REMOVED***
function NullLiteral() {
  return nullLiteralTypeAnnotation();
***REMOVED***
function RegExpLiteral() {
  return genericTypeAnnotation(identifier("RegExp"));
***REMOVED***
function ObjectExpression() {
  return genericTypeAnnotation(identifier("Object"));
***REMOVED***
function ArrayExpression() {
  return genericTypeAnnotation(identifier("Array"));
***REMOVED***
function RestElement() {
  return ArrayExpression();
***REMOVED***
RestElement.validParent = true;
function Func() {
  return genericTypeAnnotation(identifier("Function"));
***REMOVED***
const isArrayFrom = buildMatchMemberExpression("Array.from");
const isObjectKeys = buildMatchMemberExpression("Object.keys");
const isObjectValues = buildMatchMemberExpression("Object.values");
const isObjectEntries = buildMatchMemberExpression("Object.entries");
function CallExpression() {
  const {
    callee
  ***REMOVED*** = this.node;
  if (isObjectKeys(callee)) {
    return arrayTypeAnnotation(stringTypeAnnotation());
  ***REMOVED*** else if (isArrayFrom(callee) || isObjectValues(callee) || isIdentifier(callee, {
    name: "Array"
  ***REMOVED***)) {
    return arrayTypeAnnotation(anyTypeAnnotation());
  ***REMOVED*** else if (isObjectEntries(callee)) {
    return arrayTypeAnnotation(tupleTypeAnnotation([stringTypeAnnotation(), anyTypeAnnotation()]));
  ***REMOVED***
  return resolveCall(this.get("callee"));
***REMOVED***
function TaggedTemplateExpression() {
  return resolveCall(this.get("tag"));
***REMOVED***
function resolveCall(callee) {
  callee = callee.resolve();
  if (callee.isFunction()) {
    const {
      node
    ***REMOVED*** = callee;
    if (node.async) {
      if (node.generator) {
        return genericTypeAnnotation(identifier("AsyncIterator"));
      ***REMOVED*** else {
        return genericTypeAnnotation(identifier("Promise"));
      ***REMOVED***
    ***REMOVED*** else {
      if (node.generator) {
        return genericTypeAnnotation(identifier("Iterator"));
      ***REMOVED*** else if (callee.node.returnType) {
        return callee.node.returnType;
      ***REMOVED*** else {***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***
