// eslint-plugin-jest/lib/rules/max-nested-describe.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

var _experimentalUtils = require("@typescript-eslint/experimental-utils");

var _utils = require("./utils");

var _default = (0, _utils.createRule)({
  name: __filename,
  meta: {
    docs: {
      category: 'Best Practices',
      description: 'Enforces a maximum depth to nested describe calls',
      recommended: false
    ***REMOVED***,
    messages: {
      exceededMaxDepth: 'Too many nested describe calls ({{ depth ***REMOVED******REMOVED***). Maximum allowed is {{ max ***REMOVED******REMOVED***.'
    ***REMOVED***,
    type: 'suggestion',
    schema: [{
      type: 'object',
      properties: {
        max: {
          type: 'integer',
          minimum: 0
        ***REMOVED***
      ***REMOVED***,
      additionalProperties: false
    ***REMOVED***]
  ***REMOVED***,
  defaultOptions: [{
    max: 5
  ***REMOVED***],

  create(context, [{
    max
  ***REMOVED***]) {
    const describeCallbackStack = [];

    function pushDescribeCallback(node) {
      const {
        parent
      ***REMOVED*** = node;

      if ((parent === null || parent === void 0 ? void 0 : parent.type) !== _experimentalUtils.AST_NODE_TYPES.CallExpression || !(0, _utils.isDescribeCall)(parent)) {
        return;
      ***REMOVED***

      describeCallbackStack.push(0);

      if (describeCallbackStack.length > max) {
        context.report({
          node: parent,
          messageId: 'exceededMaxDepth',
          data: {
            depth: describeCallbackStack.length,
            max
          ***REMOVED***
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***

    function popDescribeCallback(node) {
      const {
        parent
      ***REMOVED*** = node;

      if ((parent === null || parent === void 0 ? void 0 : parent.type) === _experimentalUtils.AST_NODE_TYPES.CallExpression && (0, _utils.isDescribeCall)(parent)) {
        describeCallbackStack.pop();
      ***REMOVED***
    ***REMOVED***

    return {
      FunctionExpression: pushDescribeCallback,
      'FunctionExpression:exit': popDescribeCallback,
      ArrowFunctionExpression: pushDescribeCallback,
      'ArrowFunctionExpression:exit': popDescribeCallback
    ***REMOVED***;
  ***REMOVED***

***REMOVED***);

exports.default = _default;
