// eslint-plugin-jest/lib/rules/no-alias-methods.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

var _utils = require("./utils");

var _default = (0, _utils.createRule)({
  name: __filename,
  meta: {
    docs: {
      category: 'Best Practices',
      description: 'Disallow alias methods',
      recommended: false
    ***REMOVED***,
    messages: {
      replaceAlias: `Replace {{ alias ***REMOVED******REMOVED***() with its canonical name of {{ canonical ***REMOVED******REMOVED***()`
    ***REMOVED***,
    fixable: 'code',
    type: 'suggestion',
    schema: []
  ***REMOVED***,
  defaultOptions: [],

  create(context) {
    // map of jest matcher aliases & their canonical names
    const methodNames = {
      toBeCalled: 'toHaveBeenCalled',
      toBeCalledTimes: 'toHaveBeenCalledTimes',
      toBeCalledWith: 'toHaveBeenCalledWith',
      lastCalledWith: 'toHaveBeenLastCalledWith',
      nthCalledWith: 'toHaveBeenNthCalledWith',
      toReturn: 'toHaveReturned',
      toReturnTimes: 'toHaveReturnedTimes',
      toReturnWith: 'toHaveReturnedWith',
      lastReturnedWith: 'toHaveLastReturnedWith',
      nthReturnedWith: 'toHaveNthReturnedWith',
      toThrowError: 'toThrow'
    ***REMOVED***;
    return {
      CallExpression(node) {
        if (!(0, _utils.isExpectCall)(node)) {
          return;
        ***REMOVED***

        const {
          matcher
        ***REMOVED*** = (0, _utils.parseExpectCall)(node);

        if (!matcher) {
          return;
        ***REMOVED***

        const alias = matcher.name;

        if (alias in methodNames) {
          const canonical = methodNames[alias];
          context.report({
            messageId: 'replaceAlias',
            data: {
              alias,
              canonical
            ***REMOVED***,
            node: matcher.node.property,
            fix: fixer => [fixer.replaceText(matcher.node.property, canonical)]
          ***REMOVED***);
        ***REMOVED***
      ***REMOVED***

    ***REMOVED***;
  ***REMOVED***

***REMOVED***);

exports.default = _default;
