// eslint-plugin-jest/lib/rules/no-done-callback.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

var _experimentalUtils = require("@typescript-eslint/experimental-utils");

var _utils = require("./utils");

const findCallbackArg = (node, isJestEach) => {
  if (isJestEach) {
    return node.arguments[1];
  ***REMOVED***

  if ((0, _utils.isHook)(node) && node.arguments.length >= 1) {
    return node.arguments[0];
  ***REMOVED***

  if ((0, _utils.isTestCaseCall)(node) && node.arguments.length >= 2) {
    return node.arguments[1];
  ***REMOVED***

  return null;
***REMOVED***;

var _default = (0, _utils.createRule)({
  name: __filename,
  meta: {
    docs: {
      category: 'Best Practices',
      description: 'Avoid using a callback in asynchronous tests and hooks',
      recommended: 'error',
      suggestion: true
    ***REMOVED***,
    messages: {
      noDoneCallback: 'Return a Promise instead of relying on callback parameter',
      suggestWrappingInPromise: 'Wrap in `new Promise({{ callback ***REMOVED******REMOVED*** => ...`',
      useAwaitInsteadOfCallback: 'Use await instead of callback in async functions'
    ***REMOVED***,
    schema: [],
    type: 'suggestion',
    hasSuggestions: true
  ***REMOVED***,
  defaultOptions: [],

  create(context) {
    return {
      CallExpression(node) {
        var _getNodeName$endsWith, _getNodeName;

        // done is the second argument for it.each, not the first
        const isJestEach = (_getNodeName$endsWith = (_getNodeName = (0, _utils.getNodeName)(node.callee)) === null || _getNodeName === void 0 ? void 0 : _getNodeName.endsWith('.each')) !== null && _getNodeName$endsWith !== void 0 ? _getNodeName$endsWith : false;

        if (isJestEach && node.callee.type !== _experimentalUtils.AST_NODE_TYPES.TaggedTemplateExpression) {
          // isJestEach but not a TaggedTemplateExpression, so this must be
          // the `jest.each([])()` syntax which this rule doesn't support due
          // to its complexity (see jest-community/eslint-plugin-jest#710)
          return;
        ***REMOVED***

        const callback = findCallbackArg(node, isJestEach);
        const callbackArgIndex = Number(isJestEach);

        if (!callback || !(0, _utils.isFunction)(callback) || callback.params.length !== 1 + callbackArgIndex) {
          return;
        ***REMOVED***

        const argument = callback.params[callbackArgIndex];

        if (argument.type !== _experimentalUtils.AST_NODE_TYPES.Identifier) {
          context.report({
            node: argument,
            messageId: 'noDoneCallback'
          ***REMOVED***);
          return;
        ***REMOVED***

        if (callback.async) {
          context.report({
            node: argument,
            messageId: 'useAwaitInsteadOfCallback'
          ***REMOVED***);
          return;
        ***REMOVED***

        context.report({
          node: argument,
          messageId: 'noDoneCallback',
          suggest: [{
            messageId: 'suggestWrappingInPromise',
            data: {
              callback: argument.name
            ***REMOVED***,

            fix(fixer) {
              const {
                body
              ***REMOVED*** = callback;
              const sourceCode = context.getSourceCode();
              const firstBodyToken = sourceCode.getFirstToken(body);
              const lastBodyToken = sourceCode.getLastToken(body);
              const tokenBeforeArgument = sourceCode.getTokenBefore(argument);
              const tokenAfterArgument = sourceCode.getTokenAfter(argument);
              /* istanbul ignore if */

              if (!firstBodyToken || !lastBodyToken || !tokenBeforeArgument || !tokenAfterArgument) {
                throw new Error(`Unexpected null when attempting to fix ${context.getFilename()***REMOVED*** - please file a github issue at https://github.com/jest-community/eslint-plugin-jest`);
              ***REMOVED***

              const argumentInParens = tokenBeforeArgument.value === '(' && tokenAfterArgument.value === ')';
              let argumentFix = fixer.replaceText(argument, '()');

              if (argumentInParens) {
                argumentFix = fixer.remove(argument);
              ***REMOVED***

              let newCallback = argument.name;

              if (argumentInParens) {
                newCallback = `(${newCallback***REMOVED***)`;
              ***REMOVED***

              let beforeReplacement = `new Promise(${newCallback***REMOVED*** => `;
              let afterReplacement = ')';
              let replaceBefore = true;

              if (body.type === _experimentalUtils.AST_NODE_TYPES.BlockStatement) {
                const keyword = 'return';
                beforeReplacement = `${keyword***REMOVED*** ${beforeReplacement***REMOVED***{`;
                afterReplacement += '***REMOVED***';
                replaceBefore = false;
              ***REMOVED***

              return [argumentFix, replaceBefore ? fixer.insertTextBefore(firstBodyToken, beforeReplacement) : fixer.insertTextAfter(firstBodyToken, beforeReplacement), fixer.insertTextAfter(lastBodyToken, afterReplacement)];
            ***REMOVED***

          ***REMOVED***]
        ***REMOVED***);
      ***REMOVED***

    ***REMOVED***;
  ***REMOVED***

***REMOVED***);

exports.default = _default;
