// eslint-plugin-jest/lib/rules/no-duplicate-hooks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

var _utils = require("./utils");

const newHookContext = () => ({
  beforeAll: 0,
  beforeEach: 0,
  afterAll: 0,
  afterEach: 0
***REMOVED***);

var _default = (0, _utils.createRule)({
  name: __filename,
  meta: {
    docs: {
      category: 'Best Practices',
      description: 'Disallow duplicate setup and teardown hooks',
      recommended: false
    ***REMOVED***,
    messages: {
      noDuplicateHook: 'Duplicate {{hook***REMOVED******REMOVED*** in describe block'
    ***REMOVED***,
    schema: [],
    type: 'suggestion'
  ***REMOVED***,
  defaultOptions: [],

  create(context) {
    const hookContexts = [newHookContext()];
    return {
      CallExpression(node) {
        if ((0, _utils.isDescribeCall)(node)) {
          hookContexts.push(newHookContext());
        ***REMOVED***

        if ((0, _utils.isHook)(node)) {
          const currentLayer = hookContexts[hookContexts.length - 1];
          currentLayer[node.callee.name] += 1;

          if (currentLayer[node.callee.name] > 1) {
            context.report({
              messageId: 'noDuplicateHook',
              data: {
                hook: node.callee.name
              ***REMOVED***,
              node
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***,

      'CallExpression:exit'(node) {
        if ((0, _utils.isDescribeCall)(node)) {
          hookContexts.pop();
        ***REMOVED***
      ***REMOVED***

    ***REMOVED***;
  ***REMOVED***

***REMOVED***);

exports.default = _default;
