// eslint-plugin-jest/lib/rules/no-export.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

var _experimentalUtils = require("@typescript-eslint/experimental-utils");

var _utils = require("./utils");

var _default = (0, _utils.createRule)({
  name: __filename,
  meta: {
    docs: {
      category: 'Best Practices',
      description: 'Disallow using `exports` in files containing tests',
      recommended: 'error'
    ***REMOVED***,
    messages: {
      unexpectedExport: `Do not export from a test file.`
    ***REMOVED***,
    type: 'suggestion',
    schema: []
  ***REMOVED***,
  defaultOptions: [],

  create(context) {
    const exportNodes = [];
    let hasTestCase = false;
    return {
      'Program:exit'() {
        if (hasTestCase && exportNodes.length > 0) {
          for (const node of exportNodes) {
            context.report({
              node,
              messageId: 'unexpectedExport'
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***,

      CallExpression(node) {
        if ((0, _utils.isTestCaseCall)(node)) {
          hasTestCase = true;
        ***REMOVED***
      ***REMOVED***,

      'ExportNamedDeclaration, ExportDefaultDeclaration'(node) {
        exportNodes.push(node);
      ***REMOVED***,

      'AssignmentExpression > MemberExpression'(node) {
        let {
          object,
          property
        ***REMOVED*** = node;

        if (object.type === _experimentalUtils.AST_NODE_TYPES.MemberExpression) {
          ({
            object,
            property
          ***REMOVED*** = object);
        ***REMOVED***

        if ('name' in object && object.name === 'module' && property.type === _experimentalUtils.AST_NODE_TYPES.Identifier && /^exports?$/u.test(property.name)) {
          exportNodes.push(node);
        ***REMOVED***
      ***REMOVED***

    ***REMOVED***;
  ***REMOVED***

***REMOVED***);

exports.default = _default;
