// eslint-plugin-jest/lib/rules/no-hooks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

var _utils = require("./utils");

var _default = (0, _utils.createRule)({
  name: __filename,
  meta: {
    docs: {
      category: 'Best Practices',
      description: 'Disallow setup and teardown hooks',
      recommended: false
    ***REMOVED***,
    messages: {
      unexpectedHook: "Unexpected '{{ hookName ***REMOVED******REMOVED***' hook"
    ***REMOVED***,
    schema: [{
      type: 'object',
      properties: {
        allow: {
          type: 'array',
          contains: ['beforeAll', 'beforeEach', 'afterAll', 'afterEach']
        ***REMOVED***
      ***REMOVED***,
      additionalProperties: false
    ***REMOVED***],
    type: 'suggestion'
  ***REMOVED***,
  defaultOptions: [{
    allow: []
  ***REMOVED***],

  create(context, [{
    allow = []
  ***REMOVED***]) {
    return {
      CallExpression(node) {
        if ((0, _utils.isHook)(node) && !allow.includes(node.callee.name)) {
          context.report({
            node,
            messageId: 'unexpectedHook',
            data: {
              hookName: node.callee.name
            ***REMOVED***
          ***REMOVED***);
        ***REMOVED***
      ***REMOVED***

    ***REMOVED***;
  ***REMOVED***

***REMOVED***);

exports.default = _default;
