// eslint-plugin-jest/lib/rules/no-identical-title.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

var _utils = require("./utils");

const newDescribeContext = () => ({
  describeTitles: [],
  testTitles: []
***REMOVED***);

var _default = (0, _utils.createRule)({
  name: __filename,
  meta: {
    docs: {
      category: 'Best Practices',
      description: 'Disallow identical titles',
      recommended: 'error'
    ***REMOVED***,
    messages: {
      multipleTestTitle: 'Test title is used multiple times in the same describe block.',
      multipleDescribeTitle: 'Describe block title is used multiple times in the same describe block.'
    ***REMOVED***,
    schema: [],
    type: 'suggestion'
  ***REMOVED***,
  defaultOptions: [],

  create(context) {
    const contexts = [newDescribeContext()];
    return {
      CallExpression(node) {
        var _getNodeName;

        const currentLayer = contexts[contexts.length - 1];

        if ((0, _utils.isDescribeCall)(node)) {
          contexts.push(newDescribeContext());
        ***REMOVED***

        if ((_getNodeName = (0, _utils.getNodeName)(node.callee)) !== null && _getNodeName !== void 0 && _getNodeName.endsWith('.each')) {
          return;
        ***REMOVED***

        const [argument] = node.arguments;

        if (!argument || !(0, _utils.isStringNode)(argument)) {
          return;
        ***REMOVED***

        const title = (0, _utils.getStringValue)(argument);

        if ((0, _utils.isTestCaseCall)(node)) {
          if (currentLayer.testTitles.includes(title)) {
            context.report({
              messageId: 'multipleTestTitle',
              node: argument
            ***REMOVED***);
          ***REMOVED***

          currentLayer.testTitles.push(title);
        ***REMOVED***

        if (!(0, _utils.isDescribeCall)(node)) {
          return;
        ***REMOVED***

        if (currentLayer.describeTitles.includes(title)) {
          context.report({
            messageId: 'multipleDescribeTitle',
            node: argument
          ***REMOVED***);
        ***REMOVED***

        currentLayer.describeTitles.push(title);
      ***REMOVED***,

      'CallExpression:exit'(node) {
        if ((0, _utils.isDescribeCall)(node)) {
          contexts.pop();
        ***REMOVED***
      ***REMOVED***

    ***REMOVED***;
  ***REMOVED***

***REMOVED***);

exports.default = _default;
