// eslint-plugin-jest/lib/rules/no-if.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

var _experimentalUtils = require("@typescript-eslint/experimental-utils");

var _utils = require("./utils");

const testCaseNames = new Set([...Object.keys(_utils.TestCaseName), 'it.only', 'it.concurrent.only', 'it.skip', 'it.concurrent.skip', 'test.only', 'test.concurrent.only', 'test.skip', 'test.concurrent.skip', 'fit.concurrent']);

const isTestFunctionExpression = node => node.parent !== undefined && node.parent.type === _experimentalUtils.AST_NODE_TYPES.CallExpression && testCaseNames.has((0, _utils.getNodeName)(node.parent.callee));

const conditionName = {
  [_experimentalUtils.AST_NODE_TYPES.ConditionalExpression]: 'conditional',
  [_experimentalUtils.AST_NODE_TYPES.SwitchStatement]: 'switch',
  [_experimentalUtils.AST_NODE_TYPES.IfStatement]: 'if'
***REMOVED***;

var _default = (0, _utils.createRule)({
  name: __filename,
  meta: {
    docs: {
      description: 'Disallow conditional logic',
      category: 'Best Practices',
      recommended: false
    ***REMOVED***,
    messages: {
      conditionalInTest: 'Test should not contain {{ condition ***REMOVED******REMOVED*** statements.'
    ***REMOVED***,
    schema: [],
    type: 'suggestion'
  ***REMOVED***,
  defaultOptions: [],

  create(context) {
    const stack = [];

    function validate(node) {
      const lastElementInStack = stack[stack.length - 1];

      if (stack.length === 0 || !lastElementInStack) {
        return;
      ***REMOVED***

      context.report({
        data: {
          condition: conditionName[node.type]
        ***REMOVED***,
        messageId: 'conditionalInTest',
        node
      ***REMOVED***);
    ***REMOVED***

    return {
      CallExpression(node) {
        if ((0, _utils.isTestCaseCall)(node)) {
          stack.push(true);

          if ((0, _utils.getNodeName)(node).endsWith('each')) {
            stack.push(true);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***,

      FunctionExpression(node) {
        stack.push(isTestFunctionExpression(node));
      ***REMOVED***,

      FunctionDeclaration(node) {
        const declaredVariables = context.getDeclaredVariables(node);
        const testCallExpressions = (0, _utils.getTestCallExpressionsFromDeclaredVariables)(declaredVariables);
        stack.push(testCallExpressions.length > 0);
      ***REMOVED***,

      ArrowFunctionExpression(node) {
        stack.push(isTestFunctionExpression(node));
      ***REMOVED***,

      IfStatement: validate,
      SwitchStatement: validate,
      ConditionalExpression: validate,

      'CallExpression:exit'() {
        stack.pop();
      ***REMOVED***,

      'FunctionExpression:exit'() {
        stack.pop();
      ***REMOVED***,

      'FunctionDeclaration:exit'() {
        stack.pop();
      ***REMOVED***,

      'ArrowFunctionExpression:exit'() {
        stack.pop();
      ***REMOVED***

    ***REMOVED***;
  ***REMOVED***

***REMOVED***);

exports.default = _default;
