// eslint-plugin-jest/lib/rules/no-test-prefixes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

var _experimentalUtils = require("@typescript-eslint/experimental-utils");

var _utils = require("./utils");

var _default = (0, _utils.createRule)({
  name: __filename,
  meta: {
    docs: {
      category: 'Best Practices',
      description: 'Use `.only` and `.skip` over `f` and `x`',
      recommended: 'error'
    ***REMOVED***,
    messages: {
      usePreferredName: 'Use "{{ preferredNodeName ***REMOVED******REMOVED***" instead'
    ***REMOVED***,
    fixable: 'code',
    schema: [],
    type: 'suggestion'
  ***REMOVED***,
  defaultOptions: [],

  create(context) {
    return {
      CallExpression(node) {
        const nodeName = (0, _utils.getNodeName)(node.callee);
        if (!nodeName || !(0, _utils.isDescribeCall)(node) && !(0, _utils.isTestCaseCall)(node)) return;
        const preferredNodeName = getPreferredNodeName(nodeName);
        if (!preferredNodeName) return;
        const funcNode = node.callee.type === _experimentalUtils.AST_NODE_TYPES.TaggedTemplateExpression ? node.callee.tag : node.callee.type === _experimentalUtils.AST_NODE_TYPES.CallExpression ? node.callee.callee : node.callee;
        context.report({
          messageId: 'usePreferredName',
          node: node.callee,
          data: {
            preferredNodeName
          ***REMOVED***,

          fix(fixer) {
            return [fixer.replaceText(funcNode, preferredNodeName)];
          ***REMOVED***

        ***REMOVED***);
      ***REMOVED***

    ***REMOVED***;
  ***REMOVED***

***REMOVED***);

exports.default = _default;

function getPreferredNodeName(nodeName) {
  const firstChar = nodeName.charAt(0);
  const suffix = nodeName.endsWith('.each') ? '.each' : '';

  if (firstChar === 'f') {
    return `${nodeName.slice(1).replace('.each', '')***REMOVED***.only${suffix***REMOVED***`;
  ***REMOVED***

  if (firstChar === 'x') {
    return `${nodeName.slice(1).replace('.each', '')***REMOVED***.skip${suffix***REMOVED***`;
  ***REMOVED***

  return null;
***REMOVED***
