// eslint-plugin-jest/lib/rules/prefer-lowercase-title.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

var _utils = require("./utils");

const hasStringAsFirstArgument = node => node.arguments[0] && (0, _utils.isStringNode)(node.arguments[0]);

const findNodeNameAndArgument = node => {
  if (!((0, _utils.isTestCaseCall)(node) || (0, _utils.isDescribeCall)(node))) {
    return null;
  ***REMOVED***

  if (!hasStringAsFirstArgument(node)) {
    return null;
  ***REMOVED***

  return [(0, _utils.getNodeName)(node).split('.')[0], node.arguments[0]];
***REMOVED***;

const populateIgnores = ignore => {
  const ignores = [];

  if (ignore.includes(_utils.DescribeAlias.describe)) {
    ignores.push(...Object.keys(_utils.DescribeAlias));
  ***REMOVED***

  if (ignore.includes(_utils.TestCaseName.test)) {
    ignores.push(...Object.keys(_utils.TestCaseName).filter(k => k.endsWith(_utils.TestCaseName.test)));
  ***REMOVED***

  if (ignore.includes(_utils.TestCaseName.it)) {
    ignores.push(...Object.keys(_utils.TestCaseName).filter(k => k.endsWith(_utils.TestCaseName.it)));
  ***REMOVED***

  return ignores;
***REMOVED***;

var _default = (0, _utils.createRule)({
  name: __filename,
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Enforce lowercase test names',
      category: 'Best Practices',
      recommended: false
    ***REMOVED***,
    fixable: 'code',
    messages: {
      unexpectedLowercase: '`{{ method ***REMOVED******REMOVED***`s should begin with lowercase'
    ***REMOVED***,
    schema: [{
      type: 'object',
      properties: {
        ignore: {
          type: 'array',
          items: {
            enum: [_utils.DescribeAlias.describe, _utils.TestCaseName.test, _utils.TestCaseName.it]
          ***REMOVED***,
          additionalItems: false
        ***REMOVED***,
        allowedPrefixes: {
          type: 'array',
          items: {
            type: 'string'
          ***REMOVED***,
          additionalItems: false
        ***REMOVED***,
        ignoreTopLevelDescribe: {
          type: 'boolean',
          default: false
        ***REMOVED***
      ***REMOVED***,
      additionalProperties: false
    ***REMOVED***]
  ***REMOVED***,
  defaultOptions: [{
    ignore: [],
    allowedPrefixes: [],
    ignoreTopLevelDescribe: false
  ***REMOVED***],

  create(context, [{
    ignore = [],
    allowedPrefixes = [],
    ignoreTopLevelDescribe
  ***REMOVED***]) {
    const ignores = populateIgnores(ignore);
    let numberOfDescribeBlocks = 0;
    return {
      CallExpression(node) {
        if ((0, _utils.isDescribeCall)(node)) {
          numberOfDescribeBlocks++;

          if (ignoreTopLevelDescribe && numberOfDescribeBlocks === 1) {
            return;
          ***REMOVED***
        ***REMOVED***

        const results = findNodeNameAndArgument(node);

        if (!results) {
          return;
        ***REMOVED***

        const [name, firstArg] = results;
        const description = (0, _utils.getStringValue)(firstArg);

        if (allowedPrefixes.some(name => description.startsWith(name))) {
          return;
        ***REMOVED***

        const firstCharacter = description.charAt(0);

        if (!firstCharacter || firstCharacter === firstCharacter.toLowerCase() || ignores.includes(name)) {
          return;
        ***REMOVED***

        context.report({
          messageId: 'unexpectedLowercase',
          node: node.arguments[0],
          data: {
            method: name
          ***REMOVED***,

          fix(fixer) {
            const description = (0, _utils.getStringValue)(firstArg);
            const rangeIgnoringQuotes = [firstArg.range[0] + 1, firstArg.range[1] - 1];
            const newDescription = description.substring(0, 1).toLowerCase() + description.substring(1);
            return [fixer.replaceTextRange(rangeIgnoringQuotes, newDescription)];
          ***REMOVED***

        ***REMOVED***);
      ***REMOVED***,

      'CallExpression:exit'(node) {
        if ((0, _utils.isDescribeCall)(node)) {
          numberOfDescribeBlocks--;
        ***REMOVED***
      ***REMOVED***

    ***REMOVED***;
  ***REMOVED***

***REMOVED***);

exports.default = _default;
