// eslint-plugin-jest/lib/rules/prefer-called-with.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

var _utils = require("./utils");

var _default = (0, _utils.createRule)({
  name: __filename,
  meta: {
    docs: {
      category: 'Best Practices',
      description: 'Suggest using `toBeCalledWith()` or `toHaveBeenCalledWith()`',
      recommended: false
    ***REMOVED***,
    messages: {
      preferCalledWith: 'Prefer {{name***REMOVED******REMOVED***With(/* expected args */)'
    ***REMOVED***,
    type: 'suggestion',
    schema: []
  ***REMOVED***,
  defaultOptions: [],

  create(context) {
    return {
      CallExpression(node) {
        if (!(0, _utils.isExpectCall)(node)) {
          return;
        ***REMOVED***

        const {
          modifier,
          matcher
        ***REMOVED*** = (0, _utils.parseExpectCall)(node); // Could check resolves/rejects here but not a likely idiom.

        if (matcher && !modifier) {
          if (['toBeCalled', 'toHaveBeenCalled'].includes(matcher.name)) {
            context.report({
              data: {
                name: matcher.name
              ***REMOVED***,
              // todo: rename to 'matcherName'
              messageId: 'preferCalledWith',
              node: matcher.node.property
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***

    ***REMOVED***;
  ***REMOVED***

***REMOVED***);

exports.default = _default;
