// eslint-plugin-jest/lib/rules/prefer-spy-on.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

var _experimentalUtils = require("@typescript-eslint/experimental-utils");

var _utils = require("./utils");

const findNodeObject = node => {
  if ('object' in node) {
    return node.object;
  ***REMOVED***

  if (node.callee.type === _experimentalUtils.AST_NODE_TYPES.MemberExpression) {
    return node.callee.object;
  ***REMOVED***

  return null;
***REMOVED***;

const getJestFnCall = node => {
  if (node.type !== _experimentalUtils.AST_NODE_TYPES.CallExpression && node.type !== _experimentalUtils.AST_NODE_TYPES.MemberExpression) {
    return null;
  ***REMOVED***

  const obj = findNodeObject(node);

  if (!obj) {
    return null;
  ***REMOVED***

  if (obj.type === _experimentalUtils.AST_NODE_TYPES.Identifier) {
    return node.type === _experimentalUtils.AST_NODE_TYPES.CallExpression && (0, _utils.getNodeName)(node.callee) === 'jest.fn' ? node : null;
  ***REMOVED***

  return getJestFnCall(obj);
***REMOVED***;

var _default = (0, _utils.createRule)({
  name: __filename,
  meta: {
    docs: {
      category: 'Best Practices',
      description: 'Suggest using `jest.spyOn()`',
      recommended: false
    ***REMOVED***,
    messages: {
      useJestSpyOn: 'Use jest.spyOn() instead.'
    ***REMOVED***,
    fixable: 'code',
    schema: [],
    type: 'suggestion'
  ***REMOVED***,
  defaultOptions: [],

  create(context) {
    return {
      AssignmentExpression(node) {
        const {
          left,
          right
        ***REMOVED*** = node;
        if (left.type !== _experimentalUtils.AST_NODE_TYPES.MemberExpression) return;
        const jestFnCall = getJestFnCall(right);
        if (!jestFnCall) return;
        context.report({
          node,
          messageId: 'useJestSpyOn',

          fix(fixer) {
            const leftPropQuote = left.property.type === _experimentalUtils.AST_NODE_TYPES.Identifier ? "'" : '';
            const [arg] = jestFnCall.arguments;
            const argSource = arg && context.getSourceCode().getText(arg);
            const mockImplementation = argSource ? `.mockImplementation(${argSource***REMOVED***)` : '.mockImplementation()';
            return [fixer.insertTextBefore(left, `jest.spyOn(`), fixer.replaceTextRange([left.object.range[1], left.property.range[0]], `, ${leftPropQuote***REMOVED***`), fixer.replaceTextRange([left.property.range[1], jestFnCall.range[1]], `${leftPropQuote***REMOVED***)${mockImplementation***REMOVED***`)];
          ***REMOVED***

        ***REMOVED***);
      ***REMOVED***

    ***REMOVED***;
  ***REMOVED***

***REMOVED***);

exports.default = _default;
