// eslint-plugin-jest/lib/rules/require-top-level-describe.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = void 0;

var _utils = require("./utils");

const messages = {
  tooManyDescribes: 'There should not be more than {{ max ***REMOVED******REMOVED*** describe{{ s ***REMOVED******REMOVED*** at the top level',
  unexpectedTestCase: 'All test cases must be wrapped in a describe block.',
  unexpectedHook: 'All hooks must be wrapped in a describe block.'
***REMOVED***;

var _default = (0, _utils.createRule)({
  name: __filename,
  meta: {
    docs: {
      category: 'Best Practices',
      description: 'Require test cases and hooks to be inside a `describe` block',
      recommended: false
    ***REMOVED***,
    messages,
    type: 'suggestion',
    schema: [{
      type: 'object',
      properties: {
        maxNumberOfTopLevelDescribes: {
          type: 'number',
          minimum: 1
        ***REMOVED***
      ***REMOVED***,
      additionalProperties: false
    ***REMOVED***]
  ***REMOVED***,
  defaultOptions: [{***REMOVED***],

  create(context) {
    var _context$options$;

    const {
      maxNumberOfTopLevelDescribes = Infinity
    ***REMOVED*** = (_context$options$ = context.options[0]) !== null && _context$options$ !== void 0 ? _context$options$ : {***REMOVED***;
    let numberOfTopLevelDescribeBlocks = 0;
    let numberOfDescribeBlocks = 0;
    return {
      CallExpression(node) {
        if ((0, _utils.isDescribeCall)(node)) {
          numberOfDescribeBlocks++;

          if (numberOfDescribeBlocks === 1) {
            numberOfTopLevelDescribeBlocks++;

            if (numberOfTopLevelDescribeBlocks > maxNumberOfTopLevelDescribes) {
              context.report({
                node,
                messageId: 'tooManyDescribes',
                data: {
                  max: maxNumberOfTopLevelDescribes,
                  s: maxNumberOfTopLevelDescribes === 1 ? '' : 's'
                ***REMOVED***
              ***REMOVED***);
            ***REMOVED***
          ***REMOVED***

          return;
        ***REMOVED***

        if (numberOfDescribeBlocks === 0) {
          if ((0, _utils.isTestCaseCall)(node)) {
            context.report({
              node,
              messageId: 'unexpectedTestCase'
            ***REMOVED***);
            return;
          ***REMOVED***

          if ((0, _utils.isHook)(node)) {
            context.report({
              node,
              messageId: 'unexpectedHook'
            ***REMOVED***);
            return;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***,

      'CallExpression:exit'(node) {
        if ((0, _utils.isDescribeCall)(node)) {
          numberOfDescribeBlocks--;
        ***REMOVED***
      ***REMOVED***

    ***REMOVED***;
  ***REMOVED***

***REMOVED***);

exports.default = _default;
