// chrome-trace-event/dist/trace-event.js
"use strict";
/**
 * trace-event - A library to create a trace of your node app per
 * Google's Trace Event format:
 * // JSSTYLED
 *      https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU
 */
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.Tracer = void 0;
const stream_1 = require("stream");
function evCommon() {
    var hrtime = process.hrtime(); // [seconds, nanoseconds]
    var ts = hrtime[0] * 1000000 + Math.round(hrtime[1] / 1000); // microseconds
    return {
        ts,
        pid: process.pid,
        tid: process.pid // no meaningful tid for node.js
    ***REMOVED***;
***REMOVED***
class Tracer extends stream_1.Readable {
    constructor(opts = {***REMOVED***) {
        super();
        this.noStream = false;
        this.events = [];
        if (typeof opts !== "object") {
            throw new Error("Invalid options passed (must be an object)");
        ***REMOVED***
        if (opts.parent != null && typeof opts.parent !== "object") {
            throw new Error("Invalid option (parent) passed (must be an object)");
        ***REMOVED***
        if (opts.fields != null && typeof opts.fields !== "object") {
            throw new Error("Invalid option (fields) passed (must be an object)");
        ***REMOVED***
        if (opts.objectMode != null &&
            (opts.objectMode !== true && opts.objectMode !== false)) {
            throw new Error("Invalid option (objectsMode) passed (must be a boolean)");
        ***REMOVED***
        this.noStream = opts.noStream || false;
        this.parent = opts.parent;
        if (this.parent) {
            this.fields = Object.assign({***REMOVED***, opts.parent && opts.parent.fields);
        ***REMOVED***
        else {
            this.fields = {***REMOVED***;
        ***REMOVED***
        if (opts.fields) {
            Object.assign(this.fields, opts.fields);
        ***REMOVED***
        if (!this.fields.cat) {
            // trace-viewer *requires* `cat`, so let's have a fallback.
            this.fields.cat = "default";
        ***REMOVED***
        else if (Array.isArray(this.fields.cat)) {
            this.fields.cat = this.fields.cat.join(",");
        ***REMOVED***
        if (!this.fields.args) {
            // trace-viewer *requires* `args`, so let's have a fallback.
            this.fields.args = {***REMOVED***;
        ***REMOVED***
        if (this.parent) {
            // TODO: Not calling Readable ctor here. Does that cause probs?
            //      Probably if trying to pipe from the child.
            //      Might want a serpate TracerChild class for these guys.
            this._push = this.parent._push.bind(this.parent);
        ***REMOVED***
        else {
            this._objectMode = Boolean(opts.objectMode);
            var streamOpts = { objectMode: this._objectMode ***REMOVED***;
            if (this._objectMode) {
                this._push = this.push;
            ***REMOVED***
            else {
                this._push = this._pushString;
                streamOpts.encoding = "utf8";
            ***REMOVED***
            stream_1.Readable.call(this, streamOpts);
        ***REMOVED***
    ***REMOVED***
    /**
     * If in no streamMode in order to flush out the trace
     * you need to call flush.
     */
    flush() {
        if (this.noStream === true) {
            for (const evt of this.events) {
                this._push(evt);
            ***REMOVED***
            this._flush();
        ***REMOVED***
    ***REMOVED***
    _read(_) { ***REMOVED***
    _pushString(ev) {
        var separator = "";
        if (!this.firstPush) {
            this.push("[");
            this.firstPush = true;
        ***REMOVED***
        else {
            separator = ",\n";
        ***REMOVED***
        this.push(separator + JSON.stringify(ev), "utf8");
    ***REMOVED***
    _flush() {
        if (!this._objectMode) {
            this.push("]");
        ***REMOVED***
    ***REMOVED***
    child(fields) {
        return new Tracer({
            parent: this,
            fields: fields
        ***REMOVED***);
    ***REMOVED***
    begin(fields) {
        return this.mkEventFunc("B")(fields);
    ***REMOVED***
    end(fields) {
        return this.mkEventFunc("E")(fields);
    ***REMOVED***
    completeEvent(fields) {
        return this.mkEventFunc("X")(fields);
    ***REMOVED***
    instantEvent(fields) {
        return this.mkEventFunc("I")(fields);
    ***REMOVED***
    mkEventFunc(ph) {
        return (fields) => {
            var ev = evCommon();
            // Assign the event phase.
            ev.ph = ph;
            if (fields) {
                if (typeof fields === "string") {
                    ev.name = fields;
                ***REMOVED***
                else {
                    for (const k of Object.keys(fields)) {
                        if (k === "cat") {
                            ev.cat = fields.cat.join(",");
                        ***REMOVED***
                        else {
                            ev[k] = fields[k];
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
            if (!this.noStream) {
                this._push(ev);
            ***REMOVED***
            else {
                this.events.push(ev);
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***
exports.Tracer = Tracer;
/*
 * These correspond to the "Async events" in the Trace Events doc.
 *
 * Required fields:
 * - name
 * - id
 *
 * Optional fields:
 * - cat (array)
 * - args (object)
 * - TODO: stack fields, other optional fields?
 *
 * Dev Note: We don't explicitly assert that correct fields are
 * used for speed (premature optimization alert!).
 */
