// globby/index.js
'use strict';
const fs = require('fs');
const arrayUnion = require('array-union');
const merge2 = require('merge2');
const fastGlob = require('fast-glob');
const dirGlob = require('dir-glob');
const gitignore = require('./gitignore');
const {FilterStream, UniqueStream***REMOVED*** = require('./stream-utils');

const DEFAULT_FILTER = () => false;

const isNegative = pattern => pattern[0] === '!';

const assertPatternsInput = patterns => {
	if (!patterns.every(pattern => typeof pattern === 'string')) {
		throw new TypeError('Patterns must be a string or an array of strings');
	***REMOVED***
***REMOVED***;

const checkCwdOption = (options = {***REMOVED***) => {
	if (!options.cwd) {
		return;
	***REMOVED***

	let stat;
	try {
		stat = fs.statSync(options.cwd);
	***REMOVED*** catch {
		return;
	***REMOVED***

	if (!stat.isDirectory()) {
		throw new Error('The `cwd` option must be a path to a directory');
	***REMOVED***
***REMOVED***;

const getPathString = p => p.stats instanceof fs.Stats ? p.path : p;

const generateGlobTasks = (patterns, taskOptions) => {
	patterns = arrayUnion([].concat(patterns));
	assertPatternsInput(patterns);
	checkCwdOption(taskOptions);

	const globTasks = [];

	taskOptions = {
		ignore: [],
		expandDirectories: true,
		...taskOptions
	***REMOVED***;

	for (const [index, pattern] of patterns.entries()) {
		if (isNegative(pattern)) {
			continue;
		***REMOVED***

		const ignore = patterns
			.slice(index)
			.filter(pattern => isNegative(pattern))
			.map(pattern => pattern.slice(1));

		const options = {
			...taskOptions,
			ignore: taskOptions.ignore.concat(ignore)
		***REMOVED***;

		globTasks.push({pattern, options***REMOVED***);
	***REMOVED***

	return globTasks;
***REMOVED***;

const globDirs = (task, fn) => {
	let options = {***REMOVED***;
	if (task.options.cwd) {
		options.cwd = task.options.cwd;
	***REMOVED***

	if (Array.isArray(task.options.expandDirectories)) {
		options = {
			...options,
			files: task.options.expandDirectories
		***REMOVED***;
	***REMOVED*** else if (typeof task.options.expandDirectories === 'object') {
		options = {
			...options,
			...task.options.expandDirectories
		***REMOVED***;
	***REMOVED***

	return fn(task.pattern, options);
***REMOVED***;

const getPattern = (task, fn) => task.options.expandDirectories ? globDirs(task, fn) : [task.pattern];

const getFilterSync = options => {
	return options && options.gitignore ?
		gitignore.sync({cwd: options.cwd, ignore: options.ignore***REMOVED***) :
		DEFAULT_FILTER;
***REMOVED***;

const globToTask = task => glob => {
	const {options***REMOVED*** = task;
	if (options.ignore && Array.isArray(options.ignore) && options.expandDirectories) {
		options.ignore = dirGlob.sync(options.ignore);
	***REMOVED***

	return {
		pattern: glob,
		options
	***REMOVED***;
***REMOVED***;

module.exports = async (patterns, options) => {
	const globTasks = generateGlobTasks(patterns, options);

	const getFilter = async () => {
		return options && options.gitignore ?
			gitignore({cwd: options.cwd, ignore: options.ignore***REMOVED***) :
			DEFAULT_FILTER;
	***REMOVED***;

	const getTasks = async () => {
		const tasks = await Promise.all(globTasks.map(async task => {
			const globs = await getPattern(task, dirGlob);
			return Promise.all(globs.map(globToTask(task)));
		***REMOVED***));

		return arrayUnion(...tasks);
	***REMOVED***;

	const [filter, tasks] = await Promise.all([getFilter(), getTasks()]);
	const paths = await Promise.all(tasks.map(task => fastGlob(task.pattern, task.options)));

	return arrayUnion(...paths).filter(path_ => !filter(getPathString(path_)));
***REMOVED***;

module.exports.sync = (patterns, options) => {
	const globTasks = generateGlobTasks(patterns, options);

	const tasks = [];
	for (const task of globTasks) {
		const newTask = getPattern(task, dirGlob.sync).map(globToTask(task));
		tasks.push(...newTask);
	***REMOVED***

	const filter = getFilterSync(options);

	let matches = [];
	for (const task of tasks) {
		matches = arrayUnion(matches, fastGlob.sync(task.pattern, task.options));
	***REMOVED***

	return matches.filter(path_ => !filter(path_));
***REMOVED***;

module.exports.stream = (patterns, options) => {
	const globTasks = generateGlobTasks(patterns, options);

	const tasks = [];
	for (const task of globTasks) {
		const newTask = getPattern(task, dirGlob.sync).map(globToTask(task));
		tasks.push(...newTask);
	***REMOVED***

	const filter = getFilterSync(options);
	const filterStream = new FilterStream(p => !filter(p));
	const uniqueStream = new UniqueStream();

	return merge2(tasks.map(task => fastGlob.stream(task.pattern, task.options)))
		.pipe(filterStream)
		.pipe(uniqueStream);
***REMOVED***;

module.exports.generateGlobTasks = generateGlobTasks;

module.exports.hasMagic = (patterns, options) => []
	.concat(patterns)
	.some(pattern => fastGlob.isDynamicPattern(pattern, options));

module.exports.gitignore = gitignore;
