// eslint/lib/rule-tester/rule-tester.js
/**
 * @fileoverview Mocha test wrapper
 * @author Ilya Volodin
 */
"use strict";

/* global describe, it */

/*
 * This is a wrapper around mocha to allow for DRY unittests for eslint
 * Format:
 * RuleTester.run("{ruleName***REMOVED***", {
 *      valid: [
 *          "{code***REMOVED***",
 *          { code: "{code***REMOVED***", options: {options***REMOVED***, globals: {globals***REMOVED***, parser: "{parser***REMOVED***", settings: {settings***REMOVED*** ***REMOVED***
 *      ],
 *      invalid: [
 *          { code: "{code***REMOVED***", errors: {numErrors***REMOVED*** ***REMOVED***,
 *          { code: "{code***REMOVED***", errors: ["{errorMessage***REMOVED***"] ***REMOVED***,
 *          { code: "{code***REMOVED***", options: {options***REMOVED***, globals: {globals***REMOVED***, parser: "{parser***REMOVED***", settings: {settings***REMOVED***, errors: [{ message: "{errorMessage***REMOVED***", type: "{errorNodeType***REMOVED***"***REMOVED***] ***REMOVED***
 *      ]
 *  ***REMOVED***);
 *
 * Variables:
 * {code***REMOVED*** - String that represents the code to be tested
 * {options***REMOVED*** - Arguments that are passed to the configurable rules.
 * {globals***REMOVED*** - An object representing a list of variables that are
 *             registered as globals
 * {parser***REMOVED*** - String representing the parser to use
 * {settings***REMOVED*** - An object representing global settings for all rules
 * {numErrors***REMOVED*** - If failing case doesn't need to check error message,
 *               this integer will specify how many errors should be
 *               received
 * {errorMessage***REMOVED*** - Message that is returned by the rule on failure
 * {errorNodeType***REMOVED*** - AST node type that is returned by they rule as
 *                   a cause of the failure.
 */

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const
    assert = require("assert"),
    path = require("path"),
    util = require("util"),
    merge = require("lodash.merge"),
    equal = require("fast-deep-equal"),
    Traverser = require("../../lib/shared/traverser"),
    { getRuleOptionsSchema, validate ***REMOVED*** = require("../shared/config-validator"),
    { Linter, SourceCodeFixer, interpolate ***REMOVED*** = require("../linter");

const ajv = require("../shared/ajv")({ strictDefaults: true ***REMOVED***);

const espreePath = require.resolve("espree");
const parserSymbol = Symbol.for("eslint.RuleTester.parser");

//------------------------------------------------------------------------------
// Typedefs
//------------------------------------------------------------------------------

/** @typedef {import("../shared/types").Parser***REMOVED*** Parser */

/**
 * A test case that is expected to pass lint.
 * @typedef {Object***REMOVED*** ValidTestCase
 * @property {string***REMOVED*** code Code for the test case.
 * @property {any[]***REMOVED*** [options] Options for the test case.
 * @property {{ [name: string]: any ***REMOVED******REMOVED*** [settings] Settings for the test case.
 * @property {string***REMOVED*** [filename] The fake filename for the test case. Useful for rules that make assertion about filenames.
 * @property {string***REMOVED*** [parser] The absolute path for the parser.
 * @property {{ [name: string]: any ***REMOVED******REMOVED*** [parserOptions] Options for the parser.
 * @property {{ [name: string]: "readonly" | "writable" | "off" ***REMOVED******REMOVED*** [globals] The additional global variables.
 * @property {{ [name: string]: boolean ***REMOVED******REMOVED*** [env] Environments for the test case.
 * @property {boolean***REMOVED*** [only] Run only this test case or the subset of test cases with this property.
 */

/**
 * A test case that is expected to fail lint.
 * @typedef {Object***REMOVED*** InvalidTestCase
 * @property {string***REMOVED*** code Code for the test case.
 * @property {number | Array<TestCaseError | string | RegExp>***REMOVED*** errors Expected errors.
 * @property {string | null***REMOVED*** [output] The expected code after autofixes are applied. If set to `null`, the test runner will assert that no autofix is suggested.
 * @property {any[]***REMOVED*** [options] Options for the test case.
 * @property {{ [name: string]: any ***REMOVED******REMOVED*** [settings] Settings for the test case.
 * @property {string***REMOVED*** [filename] The fake filename for the test case. Useful for rules that make assertion about filenames.
 * @property {string***REMOVED*** [parser] The absolute path for the parser.
 * @property {{ [name: string]: any ***REMOVED******REMOVED*** [parserOptions] Options for the parser.
 * @property {{ [name: string]: "readonly" | "writable" | "off" ***REMOVED******REMOVED*** [globals] The additional global variables.
 * @property {{ [name: string]: boolean ***REMOVED******REMOVED*** [env] Environments for the test case.
 * @property {boolean***REMOVED*** [only] Run only this test case or the subset of test cases with this property.
 */

/**
 * A description of a reported error used in a rule tester test.
 * @typedef {Object***REMOVED*** TestCaseError
 * @property {string | RegExp***REMOVED*** [message] Message.
 * @property {string***REMOVED*** [messageId] Message ID.
 * @property {string***REMOVED*** [type] The type of the reported AST node.
 * @property {{ [name: string]: string ***REMOVED******REMOVED*** [data] The data used to fill the message template.
 * @property {number***REMOVED*** [line] The 1-based line number of the reported start location.
 * @property {number***REMOVED*** [column] The 1-based column number of the reported start location.
 * @property {number***REMOVED*** [endLine] The 1-based line number of the reported end location.
 * @property {number***REMOVED*** [endColumn] The 1-based column number of the reported end location.
 */

//------------------------------------------------------------------------------
// Private Members
//------------------------------------------------------------------------------

/*
 * testerDefaultConfig must not be modified as it allows to reset the tester to
 * the initial default configuration
 */
const testerDefaultConfig = { rules: {***REMOVED*** ***REMOVED***;
let defaultConfig = { rules: {***REMOVED*** ***REMOVED***;

/*
 * List every parameters possible on a test case that are not related to eslint
 * configuration
 */
const RuleTesterParameters = [
    "code",
    "filename",
    "options",
    "errors",
    "output",
    "only"
];

/*
 * All allowed property names in error objects.
 */
const errorObjectParameters = new Set([
    "message",
    "messageId",
    "data",
    "type",
    "line",
    "column",
    "endLine",
    "endColumn",
    "suggestions"
]);
const friendlyErrorObjectParameterList = `[${[...errorObjectParameters].map(key => `'${key***REMOVED***'`).join(", ")***REMOVED***]`;

/*
 * All allowed property names in suggestion objects.
 */
const suggestionObjectParameters = new Set([
    "desc",
    "messageId",
    "data",
    "output"
]);
const friendlySuggestionObjectParameterList = `[${[...suggestionObjectParameters].map(key => `'${key***REMOVED***'`).join(", ")***REMOVED***]`;

const hasOwnProperty = Function.call.bind(Object.hasOwnProperty);

/**
 * Clones a given value deeply.
 * Note: This ignores `parent` property.
 * @param {any***REMOVED*** x A value to clone.
 * @returns {any***REMOVED*** A cloned value.
 */
function cloneDeeplyExcludesParent(x) {
    if (typeof x === "object" && x !== null) {
        if (Array.isArray(x)) {
            return x.map(cloneDeeplyExcludesParent);
        ***REMOVED***

        const retv = {***REMOVED***;

        for (const key in x) {
            if (key !== "parent" && hasOwnProperty(x, key)) {
                retv[key] = cloneDeeplyExcludesParent(x[key]);
            ***REMOVED***
        ***REMOVED***

        return retv;
    ***REMOVED***

    return x;
***REMOVED***

/**
 * Freezes a given value deeply.
 * @param {any***REMOVED*** x A value to freeze.
 * @returns {void***REMOVED***
 */
function freezeDeeply(x) {
    if (typeof x === "object" && x !== null) {
        if (Array.isArray(x)) {
            x.forEach(freezeDeeply);
        ***REMOVED*** else {
            for (const key in x) {
                if (key !== "parent" && hasOwnProperty(x, key)) {
                    freezeDeeply(x[key]);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        Object.freeze(x);
    ***REMOVED***
***REMOVED***

/**
 * Replace control characters by `\u00xx` form.
 * @param {string***REMOVED*** text The text to sanitize.
 * @returns {string***REMOVED*** The sanitized text.
 */
function sanitize(text) {
    return text.replace(
        /[\u0000-\u0009\u000b-\u001a]/gu, // eslint-disable-line no-control-regex
        c => `\\u${c.codePointAt(0).toString(16).padStart(4, "0")***REMOVED***`
    );
***REMOVED***

/**
 * Define `start`/`end` properties as throwing error.
 * @param {string***REMOVED*** objName Object name used for error messages.
 * @param {ASTNode***REMOVED*** node The node to define.
 * @returns {void***REMOVED***
 */
function defineStartEndAsError(objName, node) {
    Object.defineProperties(node, {
        start: {
            get() {
                throw new Error(`Use ${objName***REMOVED***.range[0] instead of ${objName***REMOVED***.start`);
            ***REMOVED***,
            configurable: true,
            enumerable: false
        ***REMOVED***,
        end: {
            get() {
                throw new Error(`Use ${objName***REMOVED***.range[1] instead of ${objName***REMOVED***.end`);
            ***REMOVED***,
            configurable: true,
            enumerable: false
        ***REMOVED***
    ***REMOVED***);
***REMOVED***


/**
 * Define `start`/`end` properties of all nodes of the given AST as throwing error.
 * @param {ASTNode***REMOVED*** ast The root node to errorize `start`/`end` properties.
 * @param {Object***REMOVED*** [visitorKeys] Visitor keys to be used for traversing the given ast.
 * @returns {void***REMOVED***
 */
function defineStartEndAsErrorInTree(ast, visitorKeys) {
    Traverser.traverse(ast, { visitorKeys, enter: defineStartEndAsError.bind(null, "node") ***REMOVED***);
    ast.tokens.forEach(defineStartEndAsError.bind(null, "token"));
    ast.comments.forEach(defineStartEndAsError.bind(null, "token"));
***REMOVED***

/**
 * Wraps the given parser in order to intercept and modify return values from the `parse` and `parseForESLint` methods, for test purposes.
 * In particular, to modify ast nodes, tokens and comments to throw on access to their `start` and `end` properties.
 * @param {Parser***REMOVED*** parser Parser object.
 * @returns {Parser***REMOVED*** Wrapped parser object.
 */
function wrapParser(parser) {

    if (typeof parser.parseForESLint === "function") {
        return {
            [parserSymbol]: parser,
            parseForESLint(...args) {
                const ret = parser.parseForESLint(...args);

                defineStartEndAsErrorInTree(ret.ast, ret.visitorKeys);
                return ret;
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***

    return {
        [parserSymbol]: parser,
        parse(...args) {
            const ast = parser.parse(...args);

            defineStartEndAsErrorInTree(ast);
            return ast;
        ***REMOVED***
    ***REMOVED***;
***REMOVED***

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

// default separators for testing
const DESCRIBE = Symbol("describe");
const IT = Symbol("it");
const IT_ONLY = Symbol("itOnly");

/**
 * This is `it` default handler if `it` don't exist.
 * @this {Mocha***REMOVED***
 * @param {string***REMOVED*** text The description of the test case.
 * @param {Function***REMOVED*** method The logic of the test case.
 * @returns {any***REMOVED*** Returned value of `method`.
 */
function itDefaultHandler(text, method) {
    try {
        return method.call(this);
    ***REMOVED*** catch (err) {
        if (err instanceof assert.AssertionError) {
            err.message += ` (${util.inspect(err.actual)***REMOVED*** ${err.operator***REMOVED*** ${util.inspect(err.expected)***REMOVED***)`;
        ***REMOVED***
        throw err;
    ***REMOVED***
***REMOVED***

/**
 * This is `describe` default handler if `describe` don't exist.
 * @this {Mocha***REMOVED***
 * @param {string***REMOVED*** text The description of the test case.
 * @param {Function***REMOVED*** method The logic of the test case.
 * @returns {any***REMOVED*** Returned value of `method`.
 */
function describeDefaultHandler(text, method) {
    return method.call(this);
***REMOVED***

class RuleTester {

    /**
     * Creates a new instance of RuleTester.
     * @param {Object***REMOVED*** [testerConfig] Optional, extra configuration for the tester
     */
    constructor(testerConfig) {

        /**
         * The configuration to use for this tester. Combination of the tester
         * configuration and the default configuration.
         * @type {Object***REMOVED***
         */
        this.testerConfig = merge(
            {***REMOVED***,
            defaultConfig,
            testerConfig,
            { rules: { "rule-tester/validate-ast": "error" ***REMOVED*** ***REMOVED***
        );

        /**
         * Rule definitions to define before tests.
         * @type {Object***REMOVED***
         */
        this.rules = {***REMOVED***;
        this.linter = new Linter();
    ***REMOVED***

    /**
     * Set the configuration to use for all future tests
     * @param {Object***REMOVED*** config the configuration to use.
     * @returns {void***REMOVED***
     */
    static setDefaultConfig(config) {
        if (typeof config !== "object") {
            throw new TypeError("RuleTester.setDefaultConfig: config must be an object");
        ***REMOVED***
        defaultConfig = config;

        // Make sure the rules object exists since it is assumed to exist later
        defaultConfig.rules = defaultConfig.rules || {***REMOVED***;
    ***REMOVED***

    /**
     * Get the current configuration used for all tests
     * @returns {Object***REMOVED*** the current configuration
     */
    static getDefaultConfig() {
        return defaultConfig;
    ***REMOVED***

    /**
     * Reset the configuration to the initial configuration of the tester removing
     * any changes made until now.
     * @returns {void***REMOVED***
     */
    static resetDefaultConfig() {
        defaultConfig = merge({***REMOVED***, testerDefaultConfig);
    ***REMOVED***


    /*
     * If people use `mocha test.js --watch` command, `describe` and `it` function
     * instances are different for each execution. So `describe` and `it` should get fresh instance
     * always.
     */
    static get describe() {
        return (
            this[DESCRIBE] ||
            (typeof describe === "function" ? describe : describeDefaultHandler)
        );
    ***REMOVED***

    static set describe(value) {
        this[DESCRIBE] = value;
    ***REMOVED***

    static get it() {
        return (
            this[IT] ||
            (typeof it === "function" ? it : itDefaultHandler)
        );
    ***REMOVED***

    static set it(value) {
        this[IT] = value;
    ***REMOVED***

    /**
     * Adds the `only` property to a test to run it in isolation.
     * @param {string | ValidTestCase | InvalidTestCase***REMOVED*** item A single test to run by itself.
     * @returns {ValidTestCase | InvalidTestCase***REMOVED*** The test with `only` set.
     */
    static only(item) {
        if (typeof item === "string") {
            return { code: item, only: true ***REMOVED***;
        ***REMOVED***

        return { ...item, only: true ***REMOVED***;
    ***REMOVED***

    static get itOnly() {
        if (typeof this[IT_ONLY] === "function") {
            return this[IT_ONLY];
        ***REMOVED***
        if (typeof this[IT] === "function" && typeof this[IT].only === "function") {
            return Function.bind.call(this[IT].only, this[IT]);
        ***REMOVED***
        if (typeof it === "function" && typeof it.only === "function") {
            return Function.bind.call(it.only, it);
        ***REMOVED***

        if (typeof this[DESCRIBE] === "function" || typeof this[IT] === "function") {
            throw new Error(
                "Set `RuleTester.itOnly` to use `only` with a custom test framework.\n" +
                "See https://eslint.org/docs/developer-guide/nodejs-api#customizing-ruletester for more."
            );
        ***REMOVED***
        if (typeof it === "function") {
            throw new Error("The current test framework does not support exclusive tests with `only`.");
        ***REMOVED***
        throw new Error("To use `only`, use RuleTester with a test framework that provides `it.only()` like Mocha.");
    ***REMOVED***

    static set itOnly(value) {
        this[IT_ONLY] = value;
    ***REMOVED***

    /**
     * Define a rule for one particular run of tests.
     * @param {string***REMOVED*** name The name of the rule to define.
     * @param {Function***REMOVED*** rule The rule definition.
     * @returns {void***REMOVED***
     */
    defineRule(name, rule) {
        this.rules[name] = rule;
    ***REMOVED***

    /**
     * Adds a new rule test to execute.
     * @param {string***REMOVED*** ruleName The name of the rule to run.
     * @param {Function***REMOVED*** rule The rule to test.
     * @param {{
     *   valid: (ValidTestCase | string)[],
     *   invalid: InvalidTestCase[]
     * ***REMOVED******REMOVED*** test The collection of tests to run.
     * @returns {void***REMOVED***
     */
    run(ruleName, rule, test) {

        const testerConfig = this.testerConfig,
            requiredScenarios = ["valid", "invalid"],
            scenarioErrors = [],
            linter = this.linter;

        if (!test || typeof test !== "object") {
            throw new TypeError(`Test Scenarios for rule ${ruleName***REMOVED*** : Could not find test scenario object`);
        ***REMOVED***

        requiredScenarios.forEach(scenarioType => {
            if (!test[scenarioType]) {
                scenarioErrors.push(`Could not find any ${scenarioType***REMOVED*** test scenarios`);
            ***REMOVED***
        ***REMOVED***);

        if (scenarioErrors.length > 0) {
            throw new Error([
                `Test Scenarios for rule ${ruleName***REMOVED*** is invalid:`
            ].concat(scenarioErrors).join("\n"));
        ***REMOVED***


        linter.defineRule(ruleName, Object.assign({***REMOVED***, rule, {

            // Create a wrapper rule that freezes the `context` properties.
            create(context) {
                freezeDeeply(context.options);
                freezeDeeply(context.settings);
                freezeDeeply(context.parserOptions);

                return (typeof rule === "function" ? rule : rule.create)(context);
            ***REMOVED***
        ***REMOVED***));

        linter.defineRules(this.rules);

        /**
         * Run the rule for the given item
         * @param {string|Object***REMOVED*** item Item to run the rule against
         * @returns {Object***REMOVED*** Eslint run result
         * @private
         */
        function runRuleForItem(item) {
            let config = merge({***REMOVED***, testerConfig),
                code, filename, output, beforeAST, afterAST;

            if (typeof item === "string") {
                code = item;
            ***REMOVED*** else {
                code = item.code;

                /*
                 * Assumes everything on the item is a config except for the
                 * parameters used by this tester
                 */
                const itemConfig = { ...item ***REMOVED***;

                for (const parameter of RuleTesterParameters) {
                    delete itemConfig[parameter];
                ***REMOVED***

                /*
                 * Create the config object from the tester config and this item
                 * specific configurations.
                 */
                config = merge(
                    config,
                    itemConfig
                );
            ***REMOVED***

            if (item.filename) {
                filename = item.filename;
            ***REMOVED***

            if (hasOwnProperty(item, "options")) {
                assert(Array.isArray(item.options), "options must be an array");
                config.rules[ruleName] = [1].concat(item.options);
            ***REMOVED*** else {
                config.rules[ruleName] = 1;
            ***REMOVED***

            const schema = getRuleOptionsSchema(rule);

            /*
             * Setup AST getters.
             * The goal is to check whether or not AST was modified when
             * running the rule under test.
             */
            linter.defineRule("rule-tester/validate-ast", () => ({
                Program(node) {
                    beforeAST = cloneDeeplyExcludesParent(node);
                ***REMOVED***,
                "Program:exit"(node) {
                    afterAST = node;
                ***REMOVED***
            ***REMOVED***));

            if (typeof config.parser === "string") {
                assert(path.isAbsolute(config.parser), "Parsers provided as strings to RuleTester must be absolute paths");
            ***REMOVED*** else {
                config.parser = espreePath;
            ***REMOVED***

            linter.defineParser(config.parser, wrapParser(require(config.parser)));

            if (schema) {
                ajv.validateSchema(schema);

                if (ajv.errors) {
                    const errors = ajv.errors.map(error => {
                        const field = error.dataPath[0] === "." ? error.dataPath.slice(1) : error.dataPath;

                        return `\t${field***REMOVED***: ${error.message***REMOVED***`;
                    ***REMOVED***).join("\n");

                    throw new Error([`Schema for rule ${ruleName***REMOVED*** is invalid:`, errors]);
                ***REMOVED***

                /*
                 * `ajv.validateSchema` checks for errors in the structure of the schema (by comparing the schema against a "meta-schema"),
                 * and it reports those errors individually. However, there are other types of schema errors that only occur when compiling
                 * the schema (e.g. using invalid defaults in a schema), and only one of these errors can be reported at a time. As a result,
                 * the schema is compiled here separately from checking for `validateSchema` errors.
                 */
                try {
                    ajv.compile(schema);
                ***REMOVED*** catch (err) {
                    throw new Error(`Schema for rule ${ruleName***REMOVED*** is invalid: ${err.message***REMOVED***`);
                ***REMOVED***
            ***REMOVED***

            validate(config, "rule-tester", id => (id === ruleName ? rule : null));

            // Verify the code.
            const messages = linter.verify(code, config, filename);
            const fatalErrorMessage = messages.find(m => m.fatal);

            assert(!fatalErrorMessage, `A fatal parsing error occurred: ${fatalErrorMessage && fatalErrorMessage.message***REMOVED***`);

            // Verify if autofix makes a syntax error or not.
            if (messages.some(m => m.fix)) {
                output = SourceCodeFixer.applyFixes(code, messages).output;
                const errorMessageInFix = linter.verify(output, config, filename).find(m => m.fatal);

                assert(!errorMessageInFix, [
                    "A fatal parsing error occurred in autofix.",
                    `Error: ${errorMessageInFix && errorMessageInFix.message***REMOVED***`,
                    "Autofix output:",
                    output
                ].join("\n"));
            ***REMOVED*** else {
                output = code;
            ***REMOVED***

            return {
                messages,
                output,
                beforeAST,
                afterAST: cloneDeeplyExcludesParent(afterAST)
            ***REMOVED***;
        ***REMOVED***

        /**
         * Check if the AST was changed
         * @param {ASTNode***REMOVED*** beforeAST AST node before running
         * @param {ASTNode***REMOVED*** afterAST AST node after running
         * @returns {void***REMOVED***
         * @private
         */
        function assertASTDidntChange(beforeAST, afterAST) {
            if (!equal(beforeAST, afterAST)) {
                assert.fail("Rule should not modify AST.");
            ***REMOVED***
        ***REMOVED***

        /**
         * Check if the template is valid or not
         * all valid cases go through this
         * @param {string|Object***REMOVED*** item Item to run the rule against
         * @returns {void***REMOVED***
         * @private
         */
        function testValidTemplate(item) {
            const result = runRuleForItem(item);
            const messages = result.messages;

            assert.strictEqual(messages.length, 0, util.format("Should have no errors but had %d: %s",
                messages.length,
                util.inspect(messages)));

            assertASTDidntChange(result.beforeAST, result.afterAST);
        ***REMOVED***

        /**
         * Asserts that the message matches its expected value. If the expected
         * value is a regular expression, it is checked against the actual
         * value.
         * @param {string***REMOVED*** actual Actual value
         * @param {string|RegExp***REMOVED*** expected Expected value
         * @returns {void***REMOVED***
         * @private
         */
        function assertMessageMatches(actual, expected) {
            if (expected instanceof RegExp) {

                // assert.js doesn't have a built-in RegExp match function
                assert.ok(
                    expected.test(actual),
                    `Expected '${actual***REMOVED***' to match ${expected***REMOVED***`
                );
            ***REMOVED*** else {
                assert.strictEqual(actual, expected);
            ***REMOVED***
        ***REMOVED***

        /**
         * Check if the template is invalid or not
         * all invalid cases go through this.
         * @param {string|Object***REMOVED*** item Item to run the rule against
         * @returns {void***REMOVED***
         * @private
         */
        function testInvalidTemplate(item) {
            assert.ok(item.errors || item.errors === 0,
                `Did not specify errors for an invalid test of ${ruleName***REMOVED***`);

            if (Array.isArray(item.errors) && item.errors.length === 0) {
                assert.fail("Invalid cases must have at least one error");
            ***REMOVED***

            const ruleHasMetaMessages = hasOwnProperty(rule, "meta") && hasOwnProperty(rule.meta, "messages");
            const friendlyIDList = ruleHasMetaMessages ? `[${Object.keys(rule.meta.messages).map(key => `'${key***REMOVED***'`).join(", ")***REMOVED***]` : null;

            const result = runRuleForItem(item);
            const messages = result.messages;

            if (typeof item.errors === "number") {

                if (item.errors === 0) {
                    assert.fail("Invalid cases must have 'error' value greater than 0");
                ***REMOVED***

                assert.strictEqual(messages.length, item.errors, util.format("Should have %d error%s but had %d: %s",
                    item.errors,
                    item.errors === 1 ? "" : "s",
                    messages.length,
                    util.inspect(messages)));
            ***REMOVED*** else {
                assert.strictEqual(
                    messages.length, item.errors.length, util.format(
                        "Should have %d error%s but had %d: %s",
                        item.errors.length,
                        item.errors.length === 1 ? "" : "s",
                        messages.length,
                        util.inspect(messages)
                    )
                );

                const hasMessageOfThisRule = messages.some(m => m.ruleId === ruleName);

                for (let i = 0, l = item.errors.length; i < l; i++) {
                    const error = item.errors[i];
                    const message = messages[i];

                    assert(hasMessageOfThisRule, "Error rule name should be the same as the name of the rule being tested");

                    if (typeof error === "string" || error instanceof RegExp) {

                        // Just an error message.
                        assertMessageMatches(message.message, error);
                    ***REMOVED*** else if (typeof error === "object" && error !== null) {

                        /*
                         * Error object.
                         * This may have a message, messageId, data, node type, line, and/or
                         * column.
                         */

                        Object.keys(error).forEach(propertyName => {
                            assert.ok(
                                errorObjectParameters.has(propertyName),
                                `Invalid error property name '${propertyName***REMOVED***'. Expected one of ${friendlyErrorObjectParameterList***REMOVED***.`
                            );
                        ***REMOVED***);

                        if (hasOwnProperty(error, "message")) {
                            assert.ok(!hasOwnProperty(error, "messageId"), "Error should not specify both 'message' and a 'messageId'.");
                            assert.ok(!hasOwnProperty(error, "data"), "Error should not specify both 'data' and 'message'.");
                            assertMessageMatches(message.message, error.message);
                        ***REMOVED*** else if (hasOwnProperty(error, "messageId")) {
                            assert.ok(
                                ruleHasMetaMessages,
                                "Error can not use 'messageId' if rule under test doesn't define 'meta.messages'."
                            );
                            if (!hasOwnProperty(rule.meta.messages, error.messageId)) {
                                assert(false, `Invalid messageId '${error.messageId***REMOVED***'. Expected one of ${friendlyIDList***REMOVED***.`);
                            ***REMOVED***
                            assert.strictEqual(
                                message.messageId,
                                error.messageId,
                                `messageId '${message.messageId***REMOVED***' does not match expected messageId '${error.messageId***REMOVED***'.`
                            );
                            if (hasOwnProperty(error, "data")) {

                                /*
                                 *  if data was provided, then directly compare the returned message to a synthetic
                                 *  interpolated message using the same message ID and data provided in the test.
                                 *  See https://github.com/eslint/eslint/issues/9890 for context.
                                 */
                                const unformattedOriginalMessage = rule.meta.messages[error.messageId];
                                const rehydratedMessage = interpolate(unformattedOriginalMessage, error.data);

                                assert.strictEqual(
                                    message.message,
                                    rehydratedMessage,
                                    `Hydrated message "${rehydratedMessage***REMOVED***" does not match "${message.message***REMOVED***"`
                                );
                            ***REMOVED***
                        ***REMOVED***

                        assert.ok(
                            hasOwnProperty(error, "data") ? hasOwnProperty(error, "messageId") : true,
                            "Error must specify 'messageId' if 'data' is used."
                        );

                        if (error.type) {
                            assert.strictEqual(message.nodeType, error.type, `Error type should be ${error.type***REMOVED***, found ${message.nodeType***REMOVED***`);
                        ***REMOVED***

                        if (hasOwnProperty(error, "line")) {
                            assert.strictEqual(message.line, error.line, `Error line should be ${error.line***REMOVED***`);
                        ***REMOVED***

                        if (hasOwnProperty(error, "column")) {
                            assert.strictEqual(message.column, error.column, `Error column should be ${error.column***REMOVED***`);
                        ***REMOVED***

                        if (hasOwnProperty(error, "endLine")) {
                            assert.strictEqual(message.endLine, error.endLine, `Error endLine should be ${error.endLine***REMOVED***`);
                        ***REMOVED***

                        if (hasOwnProperty(error, "endColumn")) {
                            assert.strictEqual(message.endColumn, error.endColumn, `Error endColumn should be ${error.endColumn***REMOVED***`);
                        ***REMOVED***

                        if (hasOwnProperty(error, "suggestions")) {

                            // Support asserting there are no suggestions
                            if (!error.suggestions || (Array.isArray(error.suggestions) && error.suggestions.length === 0)) {
                                if (Array.isArray(message.suggestions) && message.suggestions.length > 0) {
                                    assert.fail(`Error should have no suggestions on error with message: "${message.message***REMOVED***"`);
                                ***REMOVED***
                            ***REMOVED*** else {
                                assert.strictEqual(Array.isArray(message.suggestions), true, `Error should have an array of suggestions. Instead received "${message.suggestions***REMOVED***" on error with message: "${message.message***REMOVED***"`);
                                assert.strictEqual(message.suggestions.length, error.suggestions.length, `Error should have ${error.suggestions.length***REMOVED*** suggestions. Instead found ${message.suggestions.length***REMOVED*** suggestions`);

                                error.suggestions.forEach((expectedSuggestion, index) => {
                                    assert.ok(
                                        typeof expectedSuggestion === "object" && expectedSuggestion !== null,
                                        "Test suggestion in 'suggestions' array must be an object."
                                    );
                                    Object.keys(expectedSuggestion).forEach(propertyName => {
                                        assert.ok(
                                            suggestionObjectParameters.has(propertyName),
                                            `Invalid suggestion property name '${propertyName***REMOVED***'. Expected one of ${friendlySuggestionObjectParameterList***REMOVED***.`
                                        );
                                    ***REMOVED***);

                                    const actualSuggestion = message.suggestions[index];
                                    const suggestionPrefix = `Error Suggestion at index ${index***REMOVED*** :`;

                                    if (hasOwnProperty(expectedSuggestion, "desc")) {
                                        assert.ok(
                                            !hasOwnProperty(expectedSuggestion, "data"),
                                            `${suggestionPrefix***REMOVED*** Test should not specify both 'desc' and 'data'.`
                                        );
                                        assert.strictEqual(
                                            actualSuggestion.desc,
                                            expectedSuggestion.desc,
                                            `${suggestionPrefix***REMOVED*** desc should be "${expectedSuggestion.desc***REMOVED***" but got "${actualSuggestion.desc***REMOVED***" instead.`
                                        );
                                    ***REMOVED***

                                    if (hasOwnProperty(expectedSuggestion, "messageId")) {
                                        assert.ok(
                                            ruleHasMetaMessages,
                                            `${suggestionPrefix***REMOVED*** Test can not use 'messageId' if rule under test doesn't define 'meta.messages'.`
                                        );
                                        assert.ok(
                                            hasOwnProperty(rule.meta.messages, expectedSuggestion.messageId),
                                            `${suggestionPrefix***REMOVED*** Test has invalid messageId '${expectedSuggestion.messageId***REMOVED***', the rule under test allows only one of ${friendlyIDList***REMOVED***.`
                                        );
                                        assert.strictEqual(
                                            actualSuggestion.messageId,
                                            expectedSuggestion.messageId,
                                            `${suggestionPrefix***REMOVED*** messageId should be '${expectedSuggestion.messageId***REMOVED***' but got '${actualSuggestion.messageId***REMOVED***' instead.`
                                        );
                                        if (hasOwnProperty(expectedSuggestion, "data")) {
                                            const unformattedMetaMessage = rule.meta.messages[expectedSuggestion.messageId];
                                            const rehydratedDesc = interpolate(unformattedMetaMessage, expectedSuggestion.data);

                                            assert.strictEqual(
                                                actualSuggestion.desc,
                                                rehydratedDesc,
                                                `${suggestionPrefix***REMOVED*** Hydrated test desc "${rehydratedDesc***REMOVED***" does not match received desc "${actualSuggestion.desc***REMOVED***".`
                                            );
                                        ***REMOVED***
                                    ***REMOVED*** else {
                                        assert.ok(
                                            !hasOwnProperty(expectedSuggestion, "data"),
                                            `${suggestionPrefix***REMOVED*** Test must specify 'messageId' if 'data' is used.`
                                        );
                                    ***REMOVED***

                                    if (hasOwnProperty(expectedSuggestion, "output")) {
                                        const codeWithAppliedSuggestion = SourceCodeFixer.applyFixes(item.code, [actualSuggestion]).output;

                                        assert.strictEqual(codeWithAppliedSuggestion, expectedSuggestion.output, `Expected the applied suggestion fix to match the test suggestion output for suggestion at index: ${index***REMOVED*** on error with message: "${message.message***REMOVED***"`);
                                    ***REMOVED***
                                ***REMOVED***);
                            ***REMOVED***
                        ***REMOVED***
                    ***REMOVED*** else {

                        // Message was an unexpected type
                        assert.fail(`Error should be a string, object, or RegExp, but found (${util.inspect(message)***REMOVED***)`);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***

            if (hasOwnProperty(item, "output")) {
                if (item.output === null) {
                    assert.strictEqual(
                        result.output,
                        item.code,
                        "Expected no autofixes to be suggested"
                    );
                ***REMOVED*** else {
                    assert.strictEqual(result.output, item.output, "Output is incorrect.");
                ***REMOVED***
            ***REMOVED*** else {
                assert.strictEqual(
                    result.output,
                    item.code,
                    "The rule fixed the code. Please add 'output' property."
                );
            ***REMOVED***

            // Rules that produce fixes must have `meta.fixable` property.
            if (result.output !== item.code) {
                assert.ok(
                    hasOwnProperty(rule, "meta"),
                    "Fixable rules should export a `meta.fixable` property."
                );

                // Linter throws if a rule that produced a fix has `meta` but doesn't have `meta.fixable`.
            ***REMOVED***

            assertASTDidntChange(result.beforeAST, result.afterAST);
        ***REMOVED***

        /*
         * This creates a mocha test suite and pipes all supplied info through
         * one of the templates above.
         */
        RuleTester.describe(ruleName, () => {
            RuleTester.describe("valid", () => {
                test.valid.forEach(valid => {
                    RuleTester[valid.only ? "itOnly" : "it"](
                        sanitize(typeof valid === "object" ? valid.code : valid),
                        () => {
                            testValidTemplate(valid);
                        ***REMOVED***
                    );
                ***REMOVED***);
            ***REMOVED***);

            RuleTester.describe("invalid", () => {
                test.invalid.forEach(invalid => {
                    RuleTester[invalid.only ? "itOnly" : "it"](
                        sanitize(invalid.code),
                        () => {
                            testInvalidTemplate(invalid);
                        ***REMOVED***
                    );
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***
***REMOVED***

RuleTester[DESCRIBE] = RuleTester[IT] = RuleTester[IT_ONLY] = null;

module.exports = RuleTester;
