// @typescript-eslint/typescript-estree/dist/simple-traverse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***);
exports.simpleTraverse = void 0;
const visitor_keys_1 = require("@typescript-eslint/visitor-keys");
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isValidNode(x) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    return x != null && typeof x === 'object' && typeof x.type === 'string';
***REMOVED***
function getVisitorKeysForNode(allVisitorKeys, node) {
    const keys = allVisitorKeys[node.type];
    return (keys !== null && keys !== void 0 ? keys : []);
***REMOVED***
class SimpleTraverser {
    constructor(selectors, setParentPointers = false) {
        this.allVisitorKeys = visitor_keys_1.visitorKeys;
        this.selectors = selectors;
        this.setParentPointers = setParentPointers;
    ***REMOVED***
    traverse(node, parent) {
        if (!isValidNode(node)) {
            return;
        ***REMOVED***
        if (this.setParentPointers) {
            node.parent = parent;
        ***REMOVED***
        if ('enter' in this.selectors) {
            this.selectors.enter(node, parent);
        ***REMOVED***
        else if (node.type in this.selectors) {
            this.selectors[node.type](node, parent);
        ***REMOVED***
        const keys = getVisitorKeysForNode(this.allVisitorKeys, node);
        if (keys.length < 1) {
            return;
        ***REMOVED***
        for (const key of keys) {
            const childOrChildren = node[key];
            if (Array.isArray(childOrChildren)) {
                for (const child of childOrChildren) {
                    this.traverse(child, node);
                ***REMOVED***
            ***REMOVED***
            else {
                this.traverse(childOrChildren, node);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
***REMOVED***
function simpleTraverse(startingNode, options, setParentPointers = false) {
    new SimpleTraverser(options, setParentPointers).traverse(startingNode, undefined);
***REMOVED***
exports.simpleTraverse = simpleTraverse;
