// postcss-svgo/src/index.js
'use strict';
const valueParser = require('postcss-value-parser');
const { optimize ***REMOVED*** = require('svgo');
const { encode, decode ***REMOVED*** = require('./lib/url');

const PLUGIN = 'postcss-svgo';
const dataURI = /data:image\/svg\+xml(;((charset=)?utf-8|base64))?,/i;
const dataURIBase64 = /data:image\/svg\+xml;base64,/i;

// the following regex will globally match:
// \b([\w-]+)       --> a word (a sequence of one or more [alphanumeric|underscore|dash] characters; followed by
// \s*=\s*          --> an equal sign character (=) between optional whitespaces; followed by
// \\"([\S\s]+?)\\" --> any characters (including whitespaces and newlines) between literal escaped quotes (\")
const escapedQuotes = /\b([\w-]+)\s*=\s*\\"([\S\s]+?)\\"/g;

/**
 * @param {string***REMOVED*** input the SVG string
 * @param {Options***REMOVED*** opts
 * @return {{result: string, isUriEncoded: boolean***REMOVED******REMOVED*** the minification result
 */
function minifySVG(input, opts) {
  let svg = input;
  let decodedUri, isUriEncoded;
  try {
    decodedUri = decode(input);
    isUriEncoded = decodedUri !== input;
  ***REMOVED*** catch (e) {
    // Swallow exception if we cannot decode the value
    isUriEncoded = false;
  ***REMOVED***

  if (isUriEncoded) {
    svg = /** @type {string***REMOVED*** */ (decodedUri);
  ***REMOVED***

  if (opts.encode !== undefined) {
    isUriEncoded = opts.encode;
  ***REMOVED***

  // normalize all escaped quote characters from svg attributes
  // from <svg attr=\"value\"... /> to <svg attr="value"... />
  // see: https://github.com/cssnano/cssnano/issues/1194
  svg = svg.replace(escapedQuotes, '$1="$2"');

  const result = optimize(svg, opts);
  if (result.error) {
    throw new Error(result.error);
  ***REMOVED***

  return {
    result: /** @type {import('svgo').OptimizedSvg***REMOVED****/ (result).data,
    isUriEncoded,
  ***REMOVED***;
***REMOVED***

/**
 * @param {import('postcss').Declaration***REMOVED*** decl
 * @param {Options***REMOVED*** opts
 * @param {import('postcss').Result***REMOVED*** postcssResult
 * @return {void***REMOVED***
 */
function minify(decl, opts, postcssResult) {
  const parsed = valueParser(decl.value);

  const minified = parsed.walk((node) => {
    if (
      node.type !== 'function' ||
      node.value.toLowerCase() !== 'url' ||
      !node.nodes.length
    ) {
      return;
    ***REMOVED***
    let { value, quote ***REMOVED*** = /** @type {valueParser.StringNode***REMOVED*** */ (
      node.nodes[0]
    );

    let optimizedValue;

    try {
      if (dataURIBase64.test(value)) {
        const url = new URL(value);
        const base64String = `${url.protocol***REMOVED***${url.pathname***REMOVED***`.replace(
          dataURI,
          ''
        );
        const svg = Buffer.from(base64String, 'base64').toString('utf8');
        const { result ***REMOVED*** = minifySVG(svg, opts);
        const data = Buffer.from(result).toString('base64');
        optimizedValue = 'data:image/svg+xml;base64,' + data + url.hash;
      ***REMOVED*** else if (dataURI.test(value)) {
        const svg = value.replace(dataURI, '');
        const { result, isUriEncoded ***REMOVED*** = minifySVG(svg, opts);
        let data = isUriEncoded ? encode(result) : result;
        // Should always encode # otherwise we yield a broken SVG
        // in Firefox (works in Chrome however). See this issue:
        // https://github.com/cssnano/cssnano/issues/245
        data = data.replace(/#/g, '%23');
        optimizedValue = 'data:image/svg+xml;charset=utf-8,' + data;
        quote = isUriEncoded ? '"' : "'";
      ***REMOVED*** else {
        return;
      ***REMOVED***
    ***REMOVED*** catch (error) {
      decl.warn(postcssResult, `${error***REMOVED***`);
      return;
    ***REMOVED***
    node.nodes[0] = Object.assign({***REMOVED***, node.nodes[0], {
      value: optimizedValue,
      quote: quote,
      type: 'string',
      before: '',
      after: '',
    ***REMOVED***);

    return false;
  ***REMOVED***);

  decl.value = minified.toString();
***REMOVED***
/** @typedef {{encode?: boolean, plugins?: object[]***REMOVED*** & import('svgo').OptimizeOptions***REMOVED*** Options */
/**
 * @type {import('postcss').PluginCreator<Options>***REMOVED***
 * @param {Options***REMOVED*** opts
 * @return {import('postcss').Plugin***REMOVED***
 */
function pluginCreator(opts = {***REMOVED***) {
  return {
    postcssPlugin: PLUGIN,

    OnceExit(css, { result ***REMOVED***) {
      css.walkDecls((decl) => {
        if (!dataURI.test(decl.value)) {
          return;
        ***REMOVED***

        minify(decl, opts, result);
      ***REMOVED***);
    ***REMOVED***,
  ***REMOVED***;
***REMOVED***

pluginCreator.postcss = true;
module.exports = pluginCreator;
