// @babel/plugin-transform-react-jsx/lib/create-plugin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
***REMOVED***);
exports.default = createPlugin;
var _pluginSyntaxJsx = require("@babel/plugin-syntax-jsx");
var _helperPluginUtils = require("@babel/helper-plugin-utils");
var _core = require("@babel/core");
var _helperModuleImports = require("@babel/helper-module-imports");
var _helperAnnotateAsPure = require("@babel/helper-annotate-as-pure");
const DEFAULT = {
  importSource: "react",
  runtime: "automatic",
  pragma: "React.createElement",
  pragmaFrag: "React.Fragment"
***REMOVED***;
const JSX_SOURCE_ANNOTATION_REGEX = /^\s*(?:\*\s*)?@jsxImportSource\s+(\S+)\s*$/m;
const JSX_RUNTIME_ANNOTATION_REGEX = /^\s*(?:\*\s*)?@jsxRuntime\s+(\S+)\s*$/m;
const JSX_ANNOTATION_REGEX = /^\s*(?:\*\s*)?@jsx\s+(\S+)\s*$/m;
const JSX_FRAG_ANNOTATION_REGEX = /^\s*(?:\*\s*)?@jsxFrag\s+(\S+)\s*$/m;
const get = (pass, name) => pass.get(`@babel/plugin-react-jsx/${name***REMOVED***`);
const set = (pass, name, v) => pass.set(`@babel/plugin-react-jsx/${name***REMOVED***`, v);
function hasProto(node) {
  return node.properties.some(value => _core.types.isObjectProperty(value, {
    computed: false,
    shorthand: false
  ***REMOVED***) && (_core.types.isIdentifier(value.key, {
    name: "__proto__"
  ***REMOVED***) || _core.types.isStringLiteral(value.key, {
    value: "__proto__"
  ***REMOVED***)));
***REMOVED***
function createPlugin({
  name,
  development
***REMOVED***) {
  return (0, _helperPluginUtils.declare)((_, options) => {
    const {
      pure: PURE_ANNOTATION,
      throwIfNamespace = true,
      filter,
      runtime: RUNTIME_DEFAULT = development ? "automatic" : "classic",
      importSource: IMPORT_SOURCE_DEFAULT = DEFAULT.importSource,
      pragma: PRAGMA_DEFAULT = DEFAULT.pragma,
      pragmaFrag: PRAGMA_FRAG_DEFAULT = DEFAULT.pragmaFrag
    ***REMOVED*** = options;
    {
      var {
        useSpread = false,
        useBuiltIns = false
      ***REMOVED*** = options;
      if (RUNTIME_DEFAULT === "classic") {
        if (typeof useSpread !== "boolean") {
          throw new Error("transform-react-jsx currently only accepts a boolean option for " + "useSpread (defaults to false)");
        ***REMOVED***
        if (typeof useBuiltIns !== "boolean") {
          throw new Error("transform-react-jsx currently only accepts a boolean option for " + "useBuiltIns (defaults to false)");
        ***REMOVED***
        if (useSpread && useBuiltIns) {
          throw new Error("transform-react-jsx currently only accepts useBuiltIns or useSpread " + "but not both");
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
    const injectMetaPropertiesVisitor = {
      JSXOpeningElement(path, state) {
        const attributes = [];
        if (isThisAllowed(path.scope)) {
          attributes.push(_core.types.jsxAttribute(_core.types.jsxIdentifier("__self"), _core.types.jsxExpressionContainer(_core.types.thisExpression())));
        ***REMOVED***
        attributes.push(_core.types.jsxAttribute(_core.types.jsxIdentifier("__source"), _core.types.jsxExpressionContainer(makeSource(path, state))));
        path.pushContainer("attributes", attributes);
      ***REMOVED***
    ***REMOVED***;
    return {
      name,
      inherits: _pluginSyntaxJsx.default,
      visitor: {
        JSXNamespacedName(path) {
          if (throwIfNamespace) {
            throw path.buildCodeFrameError(`Namespace tags are not supported by default. React's JSX doesn't support namespace tags. \
You can set \`throwIfNamespace: false\` to bypass this warning.`);
          ***REMOVED***
        ***REMOVED***,
        JSXSpreadChild(path) {
          throw path.buildCodeFrameError("Spread children are not supported in React.");
        ***REMOVED***,
        Program: {
          enter(path, state) {
            const {
              file
            ***REMOVED*** = state;
            let runtime = RUNTIME_DEFAULT;
            let source = IMPORT_SOURCE_DEFAULT;
            let pragma = PRAGMA_DEFAULT;
            let pragmaFrag = PRAGMA_FRAG_DEFAULT;
            let sourceSet = !!options.importSource;
            let pragmaSet = !!options.pragma;
            let pragmaFragSet = !!options.pragmaFrag;
            if (file.ast.comments) {
              for (const comment of file.ast.comments) {
                const sourceMatches = JSX_SOURCE_ANNOTATION_REGEX.exec(comment.value);
                if (sourceMatches) {
                  source = sourceMatches[1];
                  sourceSet = true;
                ***REMOVED***
                const runtimeMatches = JSX_RUNTIME_ANNOTATION_REGEX.exec(comment.value);
                if (runtimeMatches) {
                  runtime = runtimeMatches[1];
                ***REMOVED***
                const jsxMatches = JSX_ANNOTATION_REGEX.exec(comment.value);
                if (jsxMatches) {
                  pragma = jsxMatches[1];
                  pragmaSet = true;
                ***REMOVED***
                const jsxFragMatches = JSX_FRAG_ANNOTATION_REGEX.exec(comment.value);
                if (jsxFragMatches) {
                  pragmaFrag = jsxFragMatches[1];
                  pragmaFragSet = true;
                ***REMOVED***
              ***REMOVED***
            ***REMOVED***
            set(state, "runtime", runtime);
            if (runtime === "classic") {
              if (sourceSet) {
                throw path.buildCodeFrameError(`importSource cannot be set when runtime is classic.`);
              ***REMOVED***
              const createElement = toMemberExpression(pragma);
              const fragment = toMemberExpression(pragmaFrag);
              set(state, "id/createElement", () => _core.types.cloneNode(createElement));
              set(state, "id/fragment", () => _core.types.cloneNode(fragment));
              set(state, "defaultPure", pragma === DEFAULT.pragma);
            ***REMOVED*** else if (runtime === "automatic") {
              if (pragmaSet || pragmaFragSet) {
                throw path.buildCodeFrameError(`pragma and pragmaFrag cannot be set when runtime is automatic.`);
              ***REMOVED***
              const define = (name, id) => set(state, name, createImportLazily(state, path, id, source));
              define("id/jsx", development ? "jsxDEV" : "jsx");
              define("id/jsxs", development ? "jsxDEV" : "jsxs");
              define("id/createElement", "createElement");
              define("id/fragment", "Fragment");
              set(state, "defaultPure", source === DEFAULT.importSource);
            ***REMOVED*** else {
              throw path.buildCodeFrameError(`Runtime must be either "classic" or "automatic".`);
            ***REMOVED***
            if (development) {
              path.traverse(injectMetaPropertiesVisitor, state);
            ***REMOVED***
          ***REMOVED***
        ***REMOVED***,
        JSXFragment: {
          exit(path, file) {
            let callExpr;
            if (get(file, "runtime") === "classic") {
              callExpr = buildCreateElementFragmentCall(path, file);
            ***REMOVED*** else {
              callExpr = buildJSXFragmentCall(path, file);
            ***REMOVED***
            path.replaceWith(_core.types.inherits(callExpr, path.node));
          ***REMOVED***
        ***REMOVED***,
        JSXElement: {
          exit(path, file) {
            let callExpr;
            if (get(file, "runtime") === "classic" || shouldUseCreateElement(path)) {
              callExpr = buildCreateElementCall(path, file);
            ***REMOVED*** else {
              callExpr = buildJSXElementCall(path, file);
            ***REMOVED***
            path.replaceWith(_core.types.inherits(callExpr, path.node));
          ***REMOVED***
        ***REMOVED***,
        JSXAttribute(path) {
          if (_core.types.isJSXElement(path.node.value)) {
            path.node.value = _core.types.jsxExpressionContainer(path.node.value);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***;
    function isDerivedClass(classPath) {
      return classPath.node.superClass !== null;
    ***REMOVED***
    function isThisAllowed(scope) {
      do {
        const {
          path
        ***REMOVED*** = scope;
        if (path.isFunctionParent() && !path.isArrowFunctionExpression()) {
          if (!path.isMethod()) {
            return true;
          ***REMOVED***
          if (path.node.kind !== "constructor") {
            return true;
          ***REMOVED***
          return !isDerivedClass(path.parentPath.parentPath);
        ***REMOVED***
        if (path.isTSModuleBlock()) {
          return false;
        ***REMOVED***
      ***REMOVED*** while (scope = scope.parent);
      return true;
    ***REMOVED***
    function call(pass, name, args) {
      const node = _core.types.callExpression(get(pass, `id/${name***REMOVED***`)(), args);
      if (PURE_ANNOTATION != null ? PURE_ANNOTATION : get(pass, "defaultPure")) (0, _helperAnnotateAsPure.default)(node);
      return node;
    ***REMOVED***
    function shouldUseCreateElement(path) {
      const openingPath = path.get("openingElement");
      const attributes = openingPath.node.attributes;
      let seenPropsSpread = false;
      for (let i = 0; i < attributes.length; i++) {
        const attr = attributes[i];
        if (seenPropsSpread && _core.types.isJSXAttribute(attr) && attr.name.name === "key") {
          return true;
        ***REMOVED*** else if (_core.types.isJSXSpreadAttribute(attr)) {
          seenPropsSpread = true;
        ***REMOVED***
      ***REMOVED***
      return false;
    ***REMOVED***
    function convertJSXIdentifier(node, parent) {
      if (_core.types.isJSXIdentifier(node)) {
        if (node.name === "this" && _core.types.isReferenced(node, parent)) {
          return _core.types.thisExpression();
        ***REMOVED*** else if (_core.types.isValidIdentifier(node.name, false)) {
          node.type = "Identifier";
          return node;
        ***REMOVED*** else {
          return _core.types.stringLiteral(node.name);
        ***REMOVED***
      ***REMOVED*** else if (_core.types.isJSXMemberExpression(node)) {
        return _core.types.memberExpression(convertJSXIdentifier(node.object, node), convertJSXIdentifier(node.property, node));
      ***REMOVED*** else if (_core.types.isJSXNamespacedName(node)) {
        return _core.types.stringLiteral(`${node.namespace.name***REMOVED***:${node.name.name***REMOVED***`);
      ***REMOVED***
      return node;
    ***REMOVED***
    function convertAttributeValue(node) {
      if (_core.types.isJSXExpressionContainer(node)) {
        return node.expression;
      ***REMOVED*** else {
        return node;
      ***REMOVED***
    ***REMOVED***
    function accumulateAttribute(array, attribute) {
      if (_core.types.isJSXSpreadAttribute(attribute.node)) {
        const arg = attribute.node.argument;
        if (_core.types.isObjectExpression(arg) && !hasProto(arg)) {
          array.push(...arg.properties);
        ***REMOVED*** else {
          array.push(_core.types.spreadElement(arg));
        ***REMOVED***
        return array;
      ***REMOVED***
      const value = convertAttributeValue(attribute.node.name.name !== "key" ? attribute.node.value || _core.types.booleanLiteral(true) : attribute.node.value);
      if (attribute.node.name.name === "key" && value === null) {
        throw attribute.buildCodeFrameError('Please provide an explicit key value. Using "key" as a shorthand for "key={true***REMOVED***" is not allowed.');
      ***REMOVED***
      if (_core.types.isStringLiteral(value) && !_core.types.isJSXExpressionContainer(attribute.node.value)) {
        var _value$extra;
        value.value = value.value.replace(/\n\s+/g, " ");
        (_value$extra = value.extra) == null || delete _value$extra.raw;
      ***REMOVED***
      if (_core.types.isJSXNamespacedName(attribute.node.name)) {
        attribute.node.name = _core.types.stringLiteral(attribute.node.name.namespace.name + ":" + attribute.node.name.name.name);
      ***REMOVED*** else if (_core.types.isValidIdentifier(attribute.node.name.name, false)) {
        attribute.node.name.type = "Identifier";
      ***REMOVED*** else {
        attribute.node.name = _core.types.stringLiteral(attribute.node.name.name);
      ***REMOVED***
      array.push(_core.types.inherits(_core.types.objectProperty(attribute.node.name, value), attribute.node));
      return array;
    ***REMOVED***
    function buildChildrenProperty(children) {
      let childrenNode;
      if (children.length === 1) {
        childrenNode = children[0];
      ***REMOVED*** else if (children.length > 1) {
        childrenNode = _core.types.arrayExpression(children);
      ***REMOVED*** else {
        return undefined;
      ***REMOVED***
      return _core.types.objectProperty(_core.types.identifier("children"), childrenNode);
    ***REMOVED***
    function buildJSXElementCall(path, file) {
      const openingPath = path.get("openingElement");
      const args = [getTag(openingPath)];
      const attribsArray = [];
      const extracted = Object.create(null);
      for (const attr of openingPath.get("attributes")) {
        if (attr.isJSXAttribute() && _core.types.isJSXIdentifier(attr.node.name)) {
          const {
            name
          ***REMOVED*** = attr.node.name;
          switch (name) {
            case "__source":
            case "__self":
              if (extracted[name]) throw sourceSelfError(path, name);
            case "key":
              {
                const keyValue = convertAttributeValue(attr.node.value);
                if (keyValue === null) {
                  throw attr.buildCodeFrameError('Please provide an explicit key value. Using "key" as a shorthand for "key={true***REMOVED***" is not allowed.');
                ***REMOVED***
                extracted[name] = keyValue;
                break;
              ***REMOVED***
            default:
              attribsArray.push(attr);
          ***REMOVED***
        ***REMOVED*** else {
          attribsArray.push(attr);
        ***REMOVED***
      ***REMOVED***
      const children = _core.types.react.buildChildren(path.node);
      let attribs;
      if (attribsArray.length || children.length) {
        attribs = buildJSXOpeningElementAttributes(attribsArray, children);
      ***REMOVED*** else {
        attribs = _core.types.objectExpression([]);
      ***REMOVED***
      args.push(attribs);
      if (development) {
        var _extracted$key;
        args.push((_extracted$key = extracted.key) != null ? _extracted$key : path.scope.buildUndefinedNode(), _core.types.booleanLiteral(children.length > 1));
        if (extracted.__source) {
          args.push(extracted.__source);
          if (extracted.__self) args.push(extracted.__self);
        ***REMOVED*** else if (extracted.__self) {
          args.push(path.scope.buildUndefinedNode(), extracted.__self);
        ***REMOVED***
      ***REMOVED*** else if (extracted.key !== undefined) {
        args.push(extracted.key);
      ***REMOVED***
      return call(file, children.length > 1 ? "jsxs" : "jsx", args);
    ***REMOVED***
    function buildJSXOpeningElementAttributes(attribs, children) {
      const props = attribs.reduce(accumulateAttribute, []);
      if ((children == null ? void 0 : children.length) > 0) {
        props.push(buildChildrenProperty(children));
      ***REMOVED***
      return _core.types.objectExpression(props);
    ***REMOVED***
    function buildJSXFragmentCall(path, file) {
      const args = [get(file, "id/fragment")()];
      const children = _core.types.react.buildChildren(path.node);
      args.push(_core.types.objectExpression(children.length > 0 ? [buildChildrenProperty(children)] : []));
      if (development) {
        args.push(path.scope.buildUndefinedNode(), _core.types.booleanLiteral(children.length > 1));
      ***REMOVED***
      return call(file, children.length > 1 ? "jsxs" : "jsx", args);
    ***REMOVED***
    function buildCreateElementFragmentCall(path, file) {
      if (filter && !filter(path.node, file)) return;
      return call(file, "createElement", [get(file, "id/fragment")(), _core.types.nullLiteral(), ..._core.types.react.buildChildren(path.node)]);
    ***REMOVED***
    function buildCreateElementCall(path, file) {
      const openingPath = path.get("openingElement");
      return call(file, "createElement", [getTag(openingPath), buildCreateElementOpeningElementAttributes(file, path, openingPath.get("attributes")), ..._core.types.react.buildChildren(path.node)]);
    ***REMOVED***
    function getTag(openingPath) {
      const tagExpr = convertJSXIdentifier(openingPath.node.name, openingPath.node);
      let tagName;
      if (_core.types.isIdentifier(tagExpr)) {
        tagName = tagExpr.name;
      ***REMOVED*** else if (_core.types.isStringLiteral(tagExpr)) {
        tagName = tagExpr.value;
      ***REMOVED***
      if (_core.types.react.isCompatTag(tagName)) {
        return _core.types.stringLiteral(tagName);
      ***REMOVED*** else {
        return tagExpr;
      ***REMOVED***
    ***REMOVED***
    function buildCreateElementOpeningElementAttributes(file, path, attribs) {
      const runtime = get(file, "runtime");
      {
        if (runtime !== "automatic") {
          const objs = [];
          const props = attribs.reduce(accumulateAttribute, []);
          if (!useSpread) {
            let start = 0;
            props.forEach((prop, i) => {
              if (_core.types.isSpreadElement(prop)) {
                if (i > start) {
                  objs.push(_core.types.objectExpression(props.slice(start, i)));
                ***REMOVED***
                objs.push(prop.argument);
                start = i + 1;
              ***REMOVED***
            ***REMOVED***);
            if (props.length > start) {
              objs.push(_core.types.objectExpression(props.slice(start)));
            ***REMOVED***
          ***REMOVED*** else if (props.length) {
            objs.push(_core.types.objectExpression(props));
          ***REMOVED***
          if (!objs.length) {
            return _core.types.nullLiteral();
          ***REMOVED***
          if (objs.length === 1) {
            if (!(_core.types.isSpreadElement(props[0]) && _core.types.isObjectExpression(props[0].argument))) {
              return objs[0];
            ***REMOVED***
          ***REMOVED***
          if (!_core.types.isObjectExpression(objs[0])) {
            objs.unshift(_core.types.objectExpression([]));
          ***REMOVED***
          const helper = useBuiltIns ? _core.types.memberExpression(_core.types.identifier("Object"), _core.types.identifier("assign")) : file.addHelper("extends");
          return _core.types.callExpression(helper, objs);
        ***REMOVED***
      ***REMOVED***
      const props = [];
      const found = Object.create(null);
      for (const attr of attribs) {
        const {
          node
        ***REMOVED*** = attr;
        const name = _core.types.isJSXAttribute(node) && _core.types.isJSXIdentifier(node.name) && node.name.name;
        if (runtime === "automatic" && (name === "__source" || name === "__self")) {
          if (found[name]) throw sourceSelfError(path, name);
          found[name] = true;
        ***REMOVED***
        accumulateAttribute(props, attr);
      ***REMOVED***
      return props.length === 1 && _core.types.isSpreadElement(props[0]) && !_core.types.isObjectExpression(props[0].argument) ? props[0].argument : props.length > 0 ? _core.types.objectExpression(props) : _core.types.nullLiteral();
    ***REMOVED***
  ***REMOVED***);
  function getSource(source, importName) {
    switch (importName) {
      case "Fragment":
        return `${source***REMOVED***/${development ? "jsx-dev-runtime" : "jsx-runtime"***REMOVED***`;
      case "jsxDEV":
        return `${source***REMOVED***/jsx-dev-runtime`;
      case "jsx":
      case "jsxs":
        return `${source***REMOVED***/jsx-runtime`;
      case "createElement":
        return source;
    ***REMOVED***
  ***REMOVED***
  function createImportLazily(pass, path, importName, source) {
    return () => {
      const actualSource = getSource(source, importName);
      if ((0, _helperModuleImports.isModule)(path)) {
        let reference = get(pass, `imports/${importName***REMOVED***`);
        if (reference) return _core.types.cloneNode(reference);
        reference = (0, _helperModuleImports.addNamed)(path, importName, actualSource, {
          importedInterop: "uncompiled",
          importPosition: "after"
        ***REMOVED***);
        set(pass, `imports/${importName***REMOVED***`, reference);
        return reference;
      ***REMOVED*** else {
        let reference = get(pass, `requires/${actualSource***REMOVED***`);
        if (reference) {
          reference = _core.types.cloneNode(reference);
        ***REMOVED*** else {
          reference = (0, _helperModuleImports.addNamespace)(path, actualSource, {
            importedInterop: "uncompiled"
          ***REMOVED***);
          set(pass, `requires/${actualSource***REMOVED***`, reference);
        ***REMOVED***
        return _core.types.memberExpression(reference, _core.types.identifier(importName));
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***
***REMOVED***
function toMemberExpression(id) {
  return id.split(".").map(name => _core.types.identifier(name)).reduce((object, property) => _core.types.memberExpression(object, property));
***REMOVED***
function makeSource(path, state) {
  const location = path.node.loc;
  if (!location) {
    return path.scope.buildUndefinedNode();
  ***REMOVED***
  if (!state.fileNameIdentifier) {
    const {
      filename = ""
    ***REMOVED*** = state;
    const fileNameIdentifier = path.scope.generateUidIdentifier("_jsxFileName");
    path.scope.getProgramParent().push({
      id: fileNameIdentifier,
      init: _core.types.stringLiteral(filename)
    ***REMOVED***);
    state.fileNameIdentifier = fileNameIdentifier;
  ***REMOVED***
  return makeTrace(_core.types.cloneNode(state.fileNameIdentifier), location.start.line, location.start.column);
***REMOVED***
function makeTrace(fileNameIdentifier, lineNumber, column0Based) {
  const fileLineLiteral = lineNumber != null ? _core.types.numericLiteral(lineNumber) : _core.types.nullLiteral();
  const fileColumnLiteral = column0Based != null ? _core.types.numericLiteral(column0Based + 1) : _core.types.nullLiteral();
  return _core.template.expression.ast`{
    fileName: ${fileNameIdentifier***REMOVED***,
    lineNumber: ${fileLineLiteral***REMOVED***,
    columnNumber: ${fileColumnLiteral***REMOVED***,
  ***REMOVED***`;
***REMOVED***
function sourceSelfError(path, name) {
  const pluginName = `transform-react-jsx-${name.slice(2)***REMOVED***`;
  return path.buildCodeFrameError(`Duplicate ${name***REMOVED*** prop found. You are most likely using the deprecated ${pluginName***REMOVED*** Babel plugin. Both __source and __self are automatically set when using the automatic runtime. Please remove transform-react-jsx-source and transform-react-jsx-self from your Babel config.`);
***REMOVED***
