// connect-history-api-fallback/lib/index.js
'use strict';

var url = require('url');

exports = module.exports = function historyApiFallback(options) {
  options = options || {***REMOVED***;
  var logger = getLogger(options);

  return function(req, res, next) {
    var headers = req.headers;
    if (req.method !== 'GET' && req.method !== 'HEAD') {
      logger(
        'Not rewriting',
        req.method,
        req.url,
        'because the method is not GET or HEAD.'
      );
      return next();
    ***REMOVED*** else if (!headers || typeof headers.accept !== 'string') {
      logger(
        'Not rewriting',
        req.method,
        req.url,
        'because the client did not send an HTTP accept header.'
      );
      return next();
    ***REMOVED*** else if (headers.accept.indexOf('application/json') === 0) {
      logger(
        'Not rewriting',
        req.method,
        req.url,
        'because the client prefers JSON.'
      );
      return next();
    ***REMOVED*** else if (!acceptsHtml(headers.accept, options)) {
      logger(
        'Not rewriting',
        req.method,
        req.url,
        'because the client does not accept HTML.'
      );
      return next();
    ***REMOVED***

    var parsedUrl = url.parse(req.url);
    var rewriteTarget;
    options.rewrites = options.rewrites || [];
    for (var i = 0; i < options.rewrites.length; i++) {
      var rewrite = options.rewrites[i];
      var match = parsedUrl.pathname.match(rewrite.from);
      if (match !== null) {
        rewriteTarget = evaluateRewriteRule(parsedUrl, match, rewrite.to, req);

        if(rewriteTarget.charAt(0) !== '/') {
          logger(
            'We recommend using an absolute path for the rewrite target.',
            'Received a non-absolute rewrite target',
            rewriteTarget,
            'for URL',
            req.url
          );
        ***REMOVED***

        logger('Rewriting', req.method, req.url, 'to', rewriteTarget);
        req.url = rewriteTarget;
        return next();
      ***REMOVED***
    ***REMOVED***

    var pathname = parsedUrl.pathname;
    if (pathname.lastIndexOf('.') > pathname.lastIndexOf('/') &&
        options.disableDotRule !== true) {
      logger(
        'Not rewriting',
        req.method,
        req.url,
        'because the path includes a dot (.) character.'
      );
      return next();
    ***REMOVED***

    rewriteTarget = options.index || '/index.html';
    logger('Rewriting', req.method, req.url, 'to', rewriteTarget);
    req.url = rewriteTarget;
    next();
  ***REMOVED***;
***REMOVED***;

function evaluateRewriteRule(parsedUrl, match, rule, req) {
  if (typeof rule === 'string') {
    return rule;
  ***REMOVED*** else if (typeof rule !== 'function') {
    throw new Error('Rewrite rule can only be of type string or function.');
  ***REMOVED***

  return rule({
    parsedUrl: parsedUrl,
    match: match,
    request: req
  ***REMOVED***);
***REMOVED***

function acceptsHtml(header, options) {
  options.htmlAcceptHeaders = options.htmlAcceptHeaders || ['text/html', '*/*'];
  for (var i = 0; i < options.htmlAcceptHeaders.length; i++) {
    if (header.indexOf(options.htmlAcceptHeaders[i]) !== -1) {
      return true;
    ***REMOVED***
  ***REMOVED***
  return false;
***REMOVED***

function getLogger(options) {
  if (options && options.logger) {
    return options.logger;
  ***REMOVED*** else if (options && options.verbose) {
    // eslint-disable-next-line no-console
    return console.log.bind(console);
  ***REMOVED***
  return function(){***REMOVED***;
***REMOVED***
