// yaml/dist/schema/yaml-1.1/int.js
'use strict';

var stringifyNumber = require('../../stringify/stringifyNumber.js');

const intIdentify = (value) => typeof value === 'bigint' || Number.isInteger(value);
function intResolve(str, offset, radix, { intAsBigInt ***REMOVED***) {
    const sign = str[0];
    if (sign === '-' || sign === '+')
        offset += 1;
    str = str.substring(offset).replace(/_/g, '');
    if (intAsBigInt) {
        switch (radix) {
            case 2:
                str = `0b${str***REMOVED***`;
                break;
            case 8:
                str = `0o${str***REMOVED***`;
                break;
            case 16:
                str = `0x${str***REMOVED***`;
                break;
        ***REMOVED***
        const n = BigInt(str);
        return sign === '-' ? BigInt(-1) * n : n;
    ***REMOVED***
    const n = parseInt(str, radix);
    return sign === '-' ? -1 * n : n;
***REMOVED***
function intStringify(node, radix, prefix) {
    const { value ***REMOVED*** = node;
    if (intIdentify(value)) {
        const str = value.toString(radix);
        return value < 0 ? '-' + prefix + str.substr(1) : prefix + str;
    ***REMOVED***
    return stringifyNumber.stringifyNumber(node);
***REMOVED***
const intBin = {
    identify: intIdentify,
    default: true,
    tag: 'tag:yaml.org,2002:int',
    format: 'BIN',
    test: /^[-+]?0b[0-1_]+$/,
    resolve: (str, _onError, opt) => intResolve(str, 2, 2, opt),
    stringify: node => intStringify(node, 2, '0b')
***REMOVED***;
const intOct = {
    identify: intIdentify,
    default: true,
    tag: 'tag:yaml.org,2002:int',
    format: 'OCT',
    test: /^[-+]?0[0-7_]+$/,
    resolve: (str, _onError, opt) => intResolve(str, 1, 8, opt),
    stringify: node => intStringify(node, 8, '0')
***REMOVED***;
const int = {
    identify: intIdentify,
    default: true,
    tag: 'tag:yaml.org,2002:int',
    test: /^[-+]?[0-9][0-9_]*$/,
    resolve: (str, _onError, opt) => intResolve(str, 0, 10, opt),
    stringify: stringifyNumber.stringifyNumber
***REMOVED***;
const intHex = {
    identify: intIdentify,
    default: true,
    tag: 'tag:yaml.org,2002:int',
    format: 'HEX',
    test: /^[-+]?0x[0-9a-fA-F_]+$/,
    resolve: (str, _onError, opt) => intResolve(str, 2, 16, opt),
    stringify: node => intStringify(node, 16, '0x')
***REMOVED***;

exports.int = int;
exports.intBin = intBin;
exports.intHex = intHex;
exports.intOct = intOct;
